{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import mygene\n",
    "import io\n",
    "import re\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import numpy as np\n",
    "from typing import Optional, List, Tuple, Union, Any\n",
    "import json\n",
    "from sparse_lmm import VariableSelection\n",
    "from sklearn.linear_model import Lasso, LogisticRegression\n",
    "from sklearn.metrics import accuracy_score, mean_squared_error\n",
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path\n",
    "\n",
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")\n",
    "\n",
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n",
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data\n",
    "\n",
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict\n",
    "\n",
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols\n",
    "\n",
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')\n",
    "\n",
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1\n",
    "\n",
    "\n",
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df\n",
    "\n",
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n",
    "\n",
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df\n",
    "\n",
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n",
    "\n",
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)\n",
    "\n",
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n",
    "\n",
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "\n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise\n",
    "\n",
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df\n",
    "\n",
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df\n",
    "\n",
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n",
    "\n",
    "def get_gene_annotation(file_path, prefixes=['^', '!', '#']):\n",
    "    \"\"\"Extract from a SOFT file the gene annotation data\"\"\"\n",
    "    gene_metadata = filter_content_by_prefix(file_path, prefixes_a=prefixes, unselect=True, source_type='file',\n",
    "                                             return_df_a=True)\n",
    "    return gene_metadata[0]\n",
    "\n",
    "def get_gene_mapping(annotation, prob_col, gene_col):\n",
    "    \"\"\"Process the gene annotation to get the mapping between gene names and gene probes.\n",
    "    \"\"\"\n",
    "    mapping_data = annotation.loc[:, [prob_col, gene_col]]\n",
    "    mapping_data = mapping_data.dropna()\n",
    "    mapping_data = mapping_data.rename(columns={gene_col: 'Gene'}).astype({'ID': 'str'})\n",
    "\n",
    "    return mapping_data\n",
    "\n",
    "def apply_gene_mapping(expression_df, mapping_df):\n",
    "    \"\"\"\n",
    "    Converts measured data about gene probes into gene expression data.\n",
    "    Handles the potential many-to-many relationship between probes and genes.\n",
    "\n",
    "    Parameters:\n",
    "    expression_df (DataFrame): A DataFrame with gene expression data, indexed by 'ID'.\n",
    "    mapping_df (DataFrame): A DataFrame mapping 'ID' to 'Gene', with 'ID' as a column.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A DataFrame with mean gene expression values, indexed by 'Gene'.\n",
    "    \"\"\"\n",
    "\n",
    "    # Define a regex pattern for splitting gene names\n",
    "    split_pattern = r';|\\|+|/{2,}|,|\\[|\\]|\\(|\\)'\n",
    "\n",
    "    # Split the 'Gene' column in 'mapping_df' using the regex pattern\n",
    "    mapping_df['Gene'] = mapping_df['Gene'].str.split(split_pattern)\n",
    "    mapping_df = mapping_df.explode('Gene')\n",
    "\n",
    "    # Set 'ID' as the index of 'mapping_df' for merging\n",
    "    mapping_df.set_index('ID', inplace=True)\n",
    "\n",
    "    # Merge 'mapping_df' with 'expression_df' by their indices\n",
    "    merged_df = mapping_df.join(expression_df)\n",
    "\n",
    "    # Group by 'Gene' and calculate the mean expression values\n",
    "    gene_expression_df = merged_df.groupby('Gene').mean().dropna()\n",
    "\n",
    "    return gene_expression_df\n",
    "\n",
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n",
    "\n",
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})\n",
    "\n",
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n",
    "\n",
    "def detect_batch_effect(X):\n",
    "    \"\"\"\n",
    "    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n",
    "\n",
    "    Args:\n",
    "    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n",
    "\n",
    "    Returns:\n",
    "    bool: True if a potential batch effect is detected, False otherwise.\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "\n",
    "    # Computing XX^T\n",
    "    XXt = np.dot(X, X.T)\n",
    "\n",
    "    # Compute the eigenvalues of XX^T\n",
    "    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n",
    "    eigen_values = sorted(eigen_values, reverse=True)\n",
    "\n",
    "    # Check for large gaps in the eigenvalues\n",
    "    for i in range(len(eigen_values) - 1):\n",
    "        gap = eigen_values[i] - eigen_values[i + 1]\n",
    "        if gap > 1 / n_samples:  # You may need to adjust this threshold\n",
    "            return True\n",
    "\n",
    "    return False\n",
    "\n",
    "def normalize_data(X_train, X_test=None):\n",
    "    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n",
    "    and optionally the test data\"\"\"\n",
    "    mean = np.mean(X_train, axis=0)\n",
    "    std = np.std(X_train, axis=0)\n",
    "\n",
    "    # Handling columns with std = 0\n",
    "    std_no_zero = np.where(std == 0, 1, std)\n",
    "\n",
    "    # Normalize X_train\n",
    "    X_train_normalized = (X_train - mean) / std_no_zero\n",
    "    # Set normalized values to 0 where std was 0\n",
    "    X_train_normalized[:, std == 0] = 0\n",
    "\n",
    "    if X_test is not None:\n",
    "        X_test_normalized = (X_test - mean) / std_no_zero\n",
    "        X_test_normalized[:, std == 0] = 0\n",
    "    else:\n",
    "        X_test_normalized = None\n",
    "\n",
    "    return X_train_normalized, X_test_normalized\n",
    "\n",
    "\n",
    "class ResidualizationRegressor:\n",
    "    def __init__(self, regression_model_constructor, params=None):\n",
    "        if params is None:\n",
    "            params = {}\n",
    "        self.regression_model = regression_model_constructor(**params)\n",
    "        self.beta_Z = None  # Coefficients for regression of Y on Z\n",
    "        self.beta_X = None  # Coefficients for regression of residual on X\n",
    "        self.neg_log_p_values = None  # Negative logarithm of p-values\n",
    "        self.p_values = None  # Actual p-values\n",
    "\n",
    "    def _reshape_data(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the data to ensure it's in the correct format (2D array).\n",
    "\n",
    "        :param data: The input data (can be 1D or 2D array).\n",
    "        :return: Reshaped 2D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 1:\n",
    "            return data.reshape(-1, 1)\n",
    "        return data\n",
    "\n",
    "    def _reshape_output(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the output data to ensure it's in the correct format (1D array).\n",
    "\n",
    "        :param data: The output data (can be 1D or 2D array).\n",
    "        :return: Reshaped 1D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 2 and data.shape[1] == 1:\n",
    "            return data.ravel()\n",
    "        return data\n",
    "\n",
    "    def fit(self, X, Y, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Y = self._reshape_data(Y)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        # Step 1: Linear regression of Y on Z\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n",
    "        Y_hat = Z_ones @ self.beta_Z\n",
    "        e_Y = Y - Y_hat  # Residual of Y\n",
    "\n",
    "        # Step 2: Regress the residual on X using the included regression model\n",
    "        self.regression_model.fit(X, e_Y)\n",
    "\n",
    "        # Obtain coefficients from the regression model\n",
    "        if hasattr(self.regression_model, 'coef_'):\n",
    "            self.beta_X = self.regression_model.coef_\n",
    "        elif hasattr(self.regression_model, 'getBeta'):\n",
    "            beta_output = self.regression_model.getBeta()\n",
    "            self.beta_X = self._reshape_output(beta_output)\n",
    "\n",
    "        # Obtain negative logarithm of p-values, if available\n",
    "        if hasattr(self.regression_model, 'getNegLogP'):\n",
    "            neg_log_p_output = self.regression_model.getNegLogP()\n",
    "            if neg_log_p_output is not None:\n",
    "                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n",
    "                self.p_values = np.exp(-self.neg_log_p_values)\n",
    "                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n",
    "                p_values_Z = np.full(Z.shape[1], np.nan)\n",
    "                self.p_values = np.concatenate((p_values_Z, self.p_values))\n",
    "\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values\n",
    "\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values\n",
    "\n",
    "\n",
    "def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n",
    "    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n",
    "    indices = np.arange(X.shape[0])\n",
    "    np.random.shuffle(indices)\n",
    "\n",
    "    fold_size = len(X) // k\n",
    "    performances = []\n",
    "\n",
    "    for i in range(k):\n",
    "        # Split data into train and test based on the current fold\n",
    "        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n",
    "        train_indices = np.setdiff1d(indices, test_indices)\n",
    "\n",
    "        X_train, X_test = X[train_indices], X[test_indices]\n",
    "        Y_train, Y_test = Y[train_indices], Y[test_indices]\n",
    "        Z_train, Z_test = Z[train_indices], Z[test_indices]\n",
    "\n",
    "        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n",
    "        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n",
    "\n",
    "        # model = model_constructor(**model_params)\n",
    "        model = ResidualizationRegressor(model_constructor, model_params)\n",
    "        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n",
    "        predictions = model.predict(normalized_X_test, normalized_Z_test)\n",
    "\n",
    "        if target_type == 'binary':\n",
    "            predictions = (predictions > 0.5).astype(int)\n",
    "            Y_test = (Y_test > 0.5).astype(int)\n",
    "            performance = accuracy_score(Y_test, predictions)\n",
    "        elif target_type == 'continuous':\n",
    "            performance = mean_squared_error(Y_test, predictions)\n",
    "\n",
    "        performances.append(performance)\n",
    "\n",
    "    cv_mean = np.mean(performances)\n",
    "    cv_std = np.std(performances)\n",
    "\n",
    "    if target_type == 'binary':\n",
    "        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n",
    "    else:\n",
    "        print(f'The cross-validation MSE is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n",
    "\n",
    "    return cv_mean, cv_std\n",
    "\n",
    "def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n",
    "                     threshold: float = 0.05, save_output: bool = True,\n",
    "                     output_dir: str = './output', model_id: int = 1) -> None:\n",
    "    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n",
    "    consists of one variable about condition and multiple variables about genetic factors.\n",
    "    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n",
    "    p-values or non-zero coefficients, depending on the availability of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    model (Any): The trained regression Model.\n",
    "    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n",
    "    trait (str): The target trait of interest.\n",
    "    condition (str): The specific condition to examine within the model.\n",
    "    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n",
    "    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n",
    "    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n",
    "    model_id (int): The index of the model, 1 or 2.\n",
    "\n",
    "    Returns:\n",
    "    None: This function does not return anything but prints and optionally saves the output.\n",
    "    \"\"\"\n",
    "    coefficients = model.get_coefficients().reshape(-1).tolist()\n",
    "    p_values = model.get_p_values()\n",
    "    if p_values is None:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients\n",
    "        })\n",
    "    else:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients,\n",
    "            'p_value': p_values.reshape(-1).tolist()\n",
    "        })\n",
    "\n",
    "    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n",
    "\n",
    "    print(f\"Effect of the condition on the target variable:\")\n",
    "    print(f\"Variable: {condition}\")\n",
    "    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n",
    "    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n",
    "    if p_values is None:\n",
    "        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n",
    "        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n",
    "        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n",
    "            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n",
    "    else:\n",
    "        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n",
    "        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n",
    "        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n",
    "        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n",
    "        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n",
    "            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n",
    "\n",
    "    print(significant_genes_sorted.to_string(index=False))\n",
    "\n",
    "    # Optionally, save this to a CSV file\n",
    "    if save_output:\n",
    "        significant_genes_sorted.to_csv(\n",
    "            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Gold standard curation: Preprocessing and single-step regression\n",
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set your preferred name\n",
    "USER = \"Jake\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/jacob/Documents'\n",
    "OUTPUT_ROOT = '/Users/jacob/Documents/Output'\n",
    "TRAIT = 'Alzheimers Disease'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Data preprocessing and selection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2. The GEO dataset\n",
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE195872', 'GSE216991', 'GSE242923']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Alzheimer's-Disease\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(\"/Users/jacob/Documents\\\\GEO\\\\Alzheimer's-Disease\\\\GSE195872\\\\GSE195872_family.soft.gz\",\n",
       " \"/Users/jacob/Documents\\\\GEO\\\\Alzheimer's-Disease\\\\GSE195872\\\\GSE195872_series_matrix.txt.gz\")"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = accession_num = \"GSE195872\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inital filtering and clinical data preprocessing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Differential gene expression in sporadic and genetic forms of Alzheimerâ€™s disease and frontotemporal dementia in brain tissue and lymphoblastoid cell lines.Â \"\n",
      "!Series_summary\t\"We have performed gene expression microarray analysis of two types of dementia, Alzheimer's disease (AD) and frontotemporal dementia (FTD), using two kind of samples, frozen brain tissue and lymphoblastoid cell lines.\"\n",
      "!Series_overall_design\t\"60 samples were included: 35 brain samples and 23 lymphoblastoid cell lines. AD group was divided in sporadic early onset and familial due to mutation in PSEN1 gene. FTD was divided in genetic by mutation in MAPT, GRN and C9orf72 genes, and sporadic with tau and TDP43 deposition. Gene expression was analyzed with Clariom D microarray (Affymetrix). Differentially expressed genes (DEG) were obtained adjusting a linear model with empirical Bayes moderation of the variance. Significance threshold was set at p-value <0.05 and fold change >1.5 and <1/1.5. In the final analysis, two samples FTD.M2.B1 and AD.MF.L3 were discarded as outliers, and thus are not included in the records.\"\n"
     ]
    }
   ],
   "source": [
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM5855442</th>\n",
       "      <th>GSM5855443</th>\n",
       "      <th>GSM5855444</th>\n",
       "      <th>GSM5855445</th>\n",
       "      <th>GSM5855446</th>\n",
       "      <th>GSM5855447</th>\n",
       "      <th>GSM5855448</th>\n",
       "      <th>GSM5855449</th>\n",
       "      <th>GSM5855450</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5855490</th>\n",
       "      <th>GSM5855491</th>\n",
       "      <th>GSM5855492</th>\n",
       "      <th>GSM5855493</th>\n",
       "      <th>GSM5855494</th>\n",
       "      <th>GSM5855495</th>\n",
       "      <th>GSM5855496</th>\n",
       "      <th>GSM5855497</th>\n",
       "      <th>GSM5855498</th>\n",
       "      <th>GSM5855499</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample group: Control</td>\n",
       "      <td>sample group: EOAD</td>\n",
       "      <td>sample group: TDP43</td>\n",
       "      <td>sample group: Tau</td>\n",
       "      <td>sample group: PSEN1</td>\n",
       "      <td>sample group: C9orf72</td>\n",
       "      <td>sample group: MAPT</td>\n",
       "      <td>sample group: Control</td>\n",
       "      <td>sample group: EOAD</td>\n",
       "      <td>...</td>\n",
       "      <td>sample group: MAPT</td>\n",
       "      <td>sample group: GRN</td>\n",
       "      <td>sample group: Control</td>\n",
       "      <td>sample group: EOAD</td>\n",
       "      <td>sample group: PSEN1</td>\n",
       "      <td>sample group: GRN</td>\n",
       "      <td>sample group: Control</td>\n",
       "      <td>sample group: EOAD</td>\n",
       "      <td>sample group: PSEN1</td>\n",
       "      <td>sample group: GRN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>tissue/cell type: Frozen prefrontal cortex</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "      <td>tissue/cell type: Immortalized lymphocyte B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                  GSM5855442  \\\n",
       "0  !Sample_characteristics_ch1                              gender: Female   \n",
       "1  !Sample_characteristics_ch1                       sample group: Control   \n",
       "2  !Sample_characteristics_ch1  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855443  \\\n",
       "0                                gender: Male   \n",
       "1                          sample group: EOAD   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855444  \\\n",
       "0                              gender: Female   \n",
       "1                         sample group: TDP43   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855445  \\\n",
       "0                                gender: Male   \n",
       "1                           sample group: Tau   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855446  \\\n",
       "0                              gender: Female   \n",
       "1                         sample group: PSEN1   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855447  \\\n",
       "0                                gender: Male   \n",
       "1                       sample group: C9orf72   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855448  \\\n",
       "0                                gender: Male   \n",
       "1                          sample group: MAPT   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855449  \\\n",
       "0                                gender: Male   \n",
       "1                       sample group: Control   \n",
       "2  tissue/cell type: Frozen prefrontal cortex   \n",
       "\n",
       "                                   GSM5855450  ...  \\\n",
       "0                              gender: Female  ...   \n",
       "1                          sample group: EOAD  ...   \n",
       "2  tissue/cell type: Frozen prefrontal cortex  ...   \n",
       "\n",
       "                                    GSM5855490  \\\n",
       "0                               gender: Female   \n",
       "1                           sample group: MAPT   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855491  \\\n",
       "0                               gender: Female   \n",
       "1                            sample group: GRN   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855492  \\\n",
       "0                               gender: Female   \n",
       "1                        sample group: Control   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855493  \\\n",
       "0                                 gender: Male   \n",
       "1                           sample group: EOAD   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855494  \\\n",
       "0                               gender: Female   \n",
       "1                          sample group: PSEN1   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855495  \\\n",
       "0                               gender: Female   \n",
       "1                            sample group: GRN   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855496  \\\n",
       "0                                 gender: Male   \n",
       "1                        sample group: Control   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855497  \\\n",
       "0                                 gender: Male   \n",
       "1                           sample group: EOAD   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855498  \\\n",
       "0                                 gender: Male   \n",
       "1                          sample group: PSEN1   \n",
       "2  tissue/cell type: Immortalized lymphocyte B   \n",
       "\n",
       "                                    GSM5855499  \n",
       "0                               gender: Female  \n",
       "1                            sample group: GRN  \n",
       "2  tissue/cell type: Immortalized lymphocyte B  \n",
       "\n",
       "[3 rows x 59 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['gender: Female', 'gender: Male'],\n",
       " 1: ['sample group: Control',\n",
       "  'sample group: EOAD',\n",
       "  'sample group: TDP43',\n",
       "  'sample group: Tau',\n",
       "  'sample group: PSEN1',\n",
       "  'sample group: C9orf72',\n",
       "  'sample group: MAPT',\n",
       "  'sample group: GRN'],\n",
       " 2: ['tissue/cell type: Frozen prefrontal cortex',\n",
       "  'tissue/cell type: Immortalized lymphocyte B']}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data. Reference prompt:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Alzheimers Disease\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Alzheimers Disease\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Differential gene expression in sporadic and genetic forms of Alzheimerâ€™s disease and frontotemporal dementia in brain tissue and lymphoblastoid cell lines.Â\\xa0\"\\n!Series_summary\\t\"We have performed gene expression microarray analysis of two types of dementia, Alzheimer\\'s disease (AD) and frontotemporal dementia (FTD), using two kind of samples, frozen brain tissue and lymphoblastoid cell lines.\"\\n!Series_overall_design\\t\"60 samples were included: 35 brain samples and 23 lymphoblastoid cell lines. AD group was divided in sporadic early onset and familial due to mutation in PSEN1 gene. FTD was divided in genetic by mutation in MAPT, GRN and C9orf72 genes, and sporadic with tau and TDP43 deposition. Gene expression was analyzed with Clariom D microarray (Affymetrix). Differentially expressed genes (DEG) were obtained adjusting a linear model with empirical Bayes moderation of the variance. Significance threshold was set at p-value <0.05 and fold change >1.5 and <1/1.5. In the final analysis, two samples FTD.M2.B1 and AD.MF.L3 were discarded as outliers, and thus are not included in the records.\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'gender: Female\\', \\'gender: Male\\'], 1: [\\'sample group: Control\\', \\'sample group: EOAD\\', \\'sample group: TDP43\\', \\'sample group: Tau\\', \\'sample group: PSEN1\\', \\'sample group: C9orf72\\', \\'sample group: MAPT\\', \\'sample group: GRN\\'], 2: [\\'tissue/cell type: Frozen prefrontal cortex\\', \\'tissue/cell type: Immortalized lymphocyte B\\']}\\n'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_trait(sample_string):\n",
    "    \"\"\"\n",
    "    Convert sample group string to a binary value representing presence of Alzheimer's Disease.\n",
    "    Presence of Alzheimer's Disease or its specific genetic mutations (like 'EOAD', 'PSEN1') is represented as 1, absence as 0.\n",
    "    Unknown or unrelated groups are converted to None.\n",
    "    \"\"\"\n",
    "    ad_related_groups = ['sample group: eoad', 'sample group: psen1']  # Add other related groups if needed\n",
    "    disease = sample_string.lower()\n",
    "    if disease in ad_related_groups:\n",
    "        return 1\n",
    "    elif disease == 'sample group: control':\n",
    "        return 0\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    gender_string = gender_string.lower()\n",
    "    if gender_string == 'gender: female':\n",
    "        return 1\n",
    "    elif gender_string == 'gender: male':\n",
    "        return 0\n",
    "    else:\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jacob\\AppData\\Local\\Temp\\ipykernel_17608\\2288027067.py:344: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "C:\\Users\\jacob\\AppData\\Local\\Temp\\ipykernel_17608\\2288027067.py:344: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5855442</th>\n",
       "      <th>GSM5855443</th>\n",
       "      <th>GSM5855444</th>\n",
       "      <th>GSM5855445</th>\n",
       "      <th>GSM5855446</th>\n",
       "      <th>GSM5855447</th>\n",
       "      <th>GSM5855448</th>\n",
       "      <th>GSM5855449</th>\n",
       "      <th>GSM5855450</th>\n",
       "      <th>GSM5855451</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5855490</th>\n",
       "      <th>GSM5855491</th>\n",
       "      <th>GSM5855492</th>\n",
       "      <th>GSM5855493</th>\n",
       "      <th>GSM5855494</th>\n",
       "      <th>GSM5855495</th>\n",
       "      <th>GSM5855496</th>\n",
       "      <th>GSM5855497</th>\n",
       "      <th>GSM5855498</th>\n",
       "      <th>GSM5855499</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Alzheimers Disease</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM5855442  GSM5855443 GSM5855444 GSM5855445  GSM5855446  \\\n",
       "Alzheimers Disease           0           1       None       None           1   \n",
       "Gender                       1           0          1          0           1   \n",
       "\n",
       "                   GSM5855447 GSM5855448  GSM5855449  GSM5855450 GSM5855451  \\\n",
       "Alzheimers Disease       None       None           0           1       None   \n",
       "Gender                      0          0           0           1          0   \n",
       "\n",
       "                    ... GSM5855490  GSM5855491 GSM5855492 GSM5855493  \\\n",
       "Alzheimers Disease  ...       None        None          0          1   \n",
       "Gender              ...          1           1          1          0   \n",
       "\n",
       "                   GSM5855494  GSM5855495  GSM5855496 GSM5855497 GSM5855498  \\\n",
       "Alzheimers Disease          1        None           0          1          1   \n",
       "Gender                      1           1           0          0          0   \n",
       "\n",
       "                    GSM5855499  \n",
       "Alzheimers Disease        None  \n",
       "Gender                       1  \n",
       "\n",
       "[2 rows x 58 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5855442</th>\n",
       "      <th>GSM5855443</th>\n",
       "      <th>GSM5855444</th>\n",
       "      <th>GSM5855445</th>\n",
       "      <th>GSM5855446</th>\n",
       "      <th>GSM5855447</th>\n",
       "      <th>GSM5855448</th>\n",
       "      <th>GSM5855449</th>\n",
       "      <th>GSM5855450</th>\n",
       "      <th>GSM5855451</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5855490</th>\n",
       "      <th>GSM5855491</th>\n",
       "      <th>GSM5855492</th>\n",
       "      <th>GSM5855493</th>\n",
       "      <th>GSM5855494</th>\n",
       "      <th>GSM5855495</th>\n",
       "      <th>GSM5855496</th>\n",
       "      <th>GSM5855497</th>\n",
       "      <th>GSM5855498</th>\n",
       "      <th>GSM5855499</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AFFX-BkGr-GC03_st</th>\n",
       "      <td>2.605167</td>\n",
       "      <td>2.461824</td>\n",
       "      <td>2.406191</td>\n",
       "      <td>2.492304</td>\n",
       "      <td>2.265330</td>\n",
       "      <td>2.334452</td>\n",
       "      <td>2.529498</td>\n",
       "      <td>2.220719</td>\n",
       "      <td>2.073709</td>\n",
       "      <td>2.286326</td>\n",
       "      <td>...</td>\n",
       "      <td>2.671258</td>\n",
       "      <td>2.704205</td>\n",
       "      <td>2.668846</td>\n",
       "      <td>2.461066</td>\n",
       "      <td>2.701479</td>\n",
       "      <td>2.533192</td>\n",
       "      <td>2.933221</td>\n",
       "      <td>2.680673</td>\n",
       "      <td>2.463570</td>\n",
       "      <td>2.619673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BkGr-GC04_st</th>\n",
       "      <td>2.383394</td>\n",
       "      <td>2.375216</td>\n",
       "      <td>2.311645</td>\n",
       "      <td>2.346333</td>\n",
       "      <td>2.248943</td>\n",
       "      <td>2.286001</td>\n",
       "      <td>2.269477</td>\n",
       "      <td>2.224211</td>\n",
       "      <td>2.104069</td>\n",
       "      <td>2.256789</td>\n",
       "      <td>...</td>\n",
       "      <td>2.611905</td>\n",
       "      <td>2.642627</td>\n",
       "      <td>2.578450</td>\n",
       "      <td>2.394123</td>\n",
       "      <td>2.524937</td>\n",
       "      <td>2.440644</td>\n",
       "      <td>2.626508</td>\n",
       "      <td>2.590390</td>\n",
       "      <td>2.486765</td>\n",
       "      <td>2.523734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BkGr-GC05_st</th>\n",
       "      <td>2.413734</td>\n",
       "      <td>2.379820</td>\n",
       "      <td>2.324516</td>\n",
       "      <td>2.369255</td>\n",
       "      <td>2.241405</td>\n",
       "      <td>2.271763</td>\n",
       "      <td>2.306545</td>\n",
       "      <td>2.205598</td>\n",
       "      <td>2.101209</td>\n",
       "      <td>2.236261</td>\n",
       "      <td>...</td>\n",
       "      <td>2.634751</td>\n",
       "      <td>2.655771</td>\n",
       "      <td>2.630272</td>\n",
       "      <td>2.445022</td>\n",
       "      <td>2.569740</td>\n",
       "      <td>2.490810</td>\n",
       "      <td>2.664660</td>\n",
       "      <td>2.628133</td>\n",
       "      <td>2.508749</td>\n",
       "      <td>2.552339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BkGr-GC06_st</th>\n",
       "      <td>2.458734</td>\n",
       "      <td>2.399517</td>\n",
       "      <td>2.344613</td>\n",
       "      <td>2.364921</td>\n",
       "      <td>2.270455</td>\n",
       "      <td>2.318263</td>\n",
       "      <td>2.341740</td>\n",
       "      <td>2.245260</td>\n",
       "      <td>2.123121</td>\n",
       "      <td>2.276898</td>\n",
       "      <td>...</td>\n",
       "      <td>2.711371</td>\n",
       "      <td>2.663128</td>\n",
       "      <td>2.627970</td>\n",
       "      <td>2.469979</td>\n",
       "      <td>2.561351</td>\n",
       "      <td>2.494557</td>\n",
       "      <td>2.679816</td>\n",
       "      <td>2.663116</td>\n",
       "      <td>2.519403</td>\n",
       "      <td>2.597704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-BkGr-GC07_st</th>\n",
       "      <td>2.473711</td>\n",
       "      <td>2.436045</td>\n",
       "      <td>2.376706</td>\n",
       "      <td>2.411106</td>\n",
       "      <td>2.267717</td>\n",
       "      <td>2.359659</td>\n",
       "      <td>2.339943</td>\n",
       "      <td>2.276328</td>\n",
       "      <td>2.138360</td>\n",
       "      <td>2.314155</td>\n",
       "      <td>...</td>\n",
       "      <td>2.735943</td>\n",
       "      <td>2.732851</td>\n",
       "      <td>2.646501</td>\n",
       "      <td>2.521692</td>\n",
       "      <td>2.613861</td>\n",
       "      <td>2.505637</td>\n",
       "      <td>2.732920</td>\n",
       "      <td>2.722793</td>\n",
       "      <td>2.580322</td>\n",
       "      <td>2.653578</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GSM5855442  GSM5855443  GSM5855444  GSM5855445  GSM5855446  \\\n",
       "ID                                                                              \n",
       "AFFX-BkGr-GC03_st    2.605167    2.461824    2.406191    2.492304    2.265330   \n",
       "AFFX-BkGr-GC04_st    2.383394    2.375216    2.311645    2.346333    2.248943   \n",
       "AFFX-BkGr-GC05_st    2.413734    2.379820    2.324516    2.369255    2.241405   \n",
       "AFFX-BkGr-GC06_st    2.458734    2.399517    2.344613    2.364921    2.270455   \n",
       "AFFX-BkGr-GC07_st    2.473711    2.436045    2.376706    2.411106    2.267717   \n",
       "\n",
       "                   GSM5855447  GSM5855448  GSM5855449  GSM5855450  GSM5855451  \\\n",
       "ID                                                                              \n",
       "AFFX-BkGr-GC03_st    2.334452    2.529498    2.220719    2.073709    2.286326   \n",
       "AFFX-BkGr-GC04_st    2.286001    2.269477    2.224211    2.104069    2.256789   \n",
       "AFFX-BkGr-GC05_st    2.271763    2.306545    2.205598    2.101209    2.236261   \n",
       "AFFX-BkGr-GC06_st    2.318263    2.341740    2.245260    2.123121    2.276898   \n",
       "AFFX-BkGr-GC07_st    2.359659    2.339943    2.276328    2.138360    2.314155   \n",
       "\n",
       "                   ...  GSM5855490  GSM5855491  GSM5855492  GSM5855493  \\\n",
       "ID                 ...                                                   \n",
       "AFFX-BkGr-GC03_st  ...    2.671258    2.704205    2.668846    2.461066   \n",
       "AFFX-BkGr-GC04_st  ...    2.611905    2.642627    2.578450    2.394123   \n",
       "AFFX-BkGr-GC05_st  ...    2.634751    2.655771    2.630272    2.445022   \n",
       "AFFX-BkGr-GC06_st  ...    2.711371    2.663128    2.627970    2.469979   \n",
       "AFFX-BkGr-GC07_st  ...    2.735943    2.732851    2.646501    2.521692   \n",
       "\n",
       "                   GSM5855494  GSM5855495  GSM5855496  GSM5855497  GSM5855498  \\\n",
       "ID                                                                              \n",
       "AFFX-BkGr-GC03_st    2.701479    2.533192    2.933221    2.680673    2.463570   \n",
       "AFFX-BkGr-GC04_st    2.524937    2.440644    2.626508    2.590390    2.486765   \n",
       "AFFX-BkGr-GC05_st    2.569740    2.490810    2.664660    2.628133    2.508749   \n",
       "AFFX-BkGr-GC06_st    2.561351    2.494557    2.679816    2.663116    2.519403   \n",
       "AFFX-BkGr-GC07_st    2.613861    2.505637    2.732920    2.722793    2.580322   \n",
       "\n",
       "                   GSM5855499  \n",
       "ID                             \n",
       "AFFX-BkGr-GC03_st    2.619673  \n",
       "AFFX-BkGr-GC04_st    2.523734  \n",
       "AFFX-BkGr-GC05_st    2.552339  \n",
       "AFFX-BkGr-GC06_st    2.597704  \n",
       "AFFX-BkGr-GC07_st    2.653578  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AFFX-BkGr-GC03_st',\n",
       " 'AFFX-BkGr-GC04_st',\n",
       " 'AFFX-BkGr-GC05_st',\n",
       " 'AFFX-BkGr-GC06_st',\n",
       " 'AFFX-BkGr-GC07_st',\n",
       " 'AFFX-BkGr-GC08_st',\n",
       " 'AFFX-BkGr-GC09_st',\n",
       " 'AFFX-BkGr-GC10_st',\n",
       " 'AFFX-BkGr-GC11_st',\n",
       " 'AFFX-BkGr-GC12_st',\n",
       " 'AFFX-BkGr-GC13_st',\n",
       " 'AFFX-BkGr-GC14_st',\n",
       " 'AFFX-BkGr-GC15_st',\n",
       " 'AFFX-BkGr-GC16_st',\n",
       " 'AFFX-BkGr-GC17_st',\n",
       " 'AFFX-BkGr-GC18_st',\n",
       " 'AFFX-BkGr-GC19_st',\n",
       " 'AFFX-BkGr-GC20_st',\n",
       " 'AFFX-BkGr-GC21_st',\n",
       " 'AFFX-BkGr-GC22_st']"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row.\n",
    "\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['AFFX-BkGr-GC03_st', 'AFFX-BkGr-GC04_st', 'AFFX-BkGr-GC05_st', 'AFFX-BkGr-GC06_st', 'AFFX-BkGr-GC07_st', 'AFFX-BkGr-GC08_st', 'AFFX-BkGr-GC09_st', 'AFFX-BkGr-GC10_st', 'AFFX-BkGr-GC11_st', 'AFFX-BkGr-GC12_st', 'AFFX-BkGr-GC13_st', 'AFFX-BkGr-GC14_st', 'AFFX-BkGr-GC15_st', 'AFFX-BkGr-GC16_st', 'AFFX-BkGr-GC17_st', 'AFFX-BkGr-GC18_st', 'AFFX-BkGr-GC19_st', 'AFFX-BkGr-GC20_st', 'AFFX-BkGr-GC21_st', 'AFFX-BkGr-GC22_st']\\n\""
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "requires_gene_mapping = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[71], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_gene_mapping:\n\u001b[1;32m----> 2\u001b[0m     gene_annotation \u001b[38;5;241m=\u001b[39m \u001b[43mget_gene_annotation\u001b[49m\u001b[43m(\u001b[49m\u001b[43msoft_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      3\u001b[0m     gene_annotation_summary \u001b[38;5;241m=\u001b[39m preview_df(gene_annotation)\n\u001b[0;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene_annotation_summary)\n",
      "Cell \u001b[1;32mIn[47], line 396\u001b[0m, in \u001b[0;36mget_gene_annotation\u001b[1;34m(file_path, prefixes)\u001b[0m\n\u001b[0;32m    394\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_gene_annotation\u001b[39m(file_path, prefixes\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m^\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    395\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Extract from a SOFT file the gene annotation data\"\"\"\u001b[39;00m\n\u001b[1;32m--> 396\u001b[0m     gene_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_content_by_prefix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprefixes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munselect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfile\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    397\u001b[0m \u001b[43m                                             \u001b[49m\u001b[43mreturn_df_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    398\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m gene_metadata[\u001b[38;5;241m0\u001b[39m]\n",
      "Cell \u001b[1;32mIn[47], line 89\u001b[0m, in \u001b[0;36mfilter_content_by_prefix\u001b[1;34m(source, prefixes_a, prefixes_b, unselect, source_type, return_df_a, return_df_b)\u001b[0m\n\u001b[0;32m     87\u001b[0m filtered_content_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(filtered_lines_a)\n\u001b[0;32m     88\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m return_df_a:\n\u001b[1;32m---> 89\u001b[0m     filtered_content_a \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfiltered_content_a\u001b[49m\u001b[43m)\u001b[49m, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, low_memory\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, on_bad_lines\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mskip\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m     90\u001b[0m filtered_content_b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m     91\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filtered_lines_b:\n",
      "\u001b[1;31mMemoryError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively. Reference prompt:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_annotation_summary' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[21], line 10\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_gene_mapping:\n\u001b[0;32m      2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'''\u001b[39m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;124m    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\u001b[39m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;124m    \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgene_row_ids\u001b[38;5;132;01m}\u001b[39;00m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;124m    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\u001b[39m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;124m    identifier_key = \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mkey_name1\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;124m    gene_symbol_key = \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mkey_name2\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      8\u001b[0m \n\u001b[0;32m      9\u001b[0m \u001b[38;5;124m    Gene annotation dictionary:\u001b[39m\n\u001b[1;32m---> 10\u001b[0m \u001b[38;5;124m    \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[43mgene_annotation_summary\u001b[49m\u001b[38;5;132;01m}\u001b[39;00m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;124m    \u001b[39m\u001b[38;5;124m'''\u001b[39m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'gene_annotation_summary' is not defined"
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GENE_SYMBOL'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['14q0_st', '14qI-1_st', '14qI-1_x_st', '14qI-2_st', '14qI-3_x_st', '14qI-4_st', '14qI-4_x_st', '14q\n",
      "1000 input query terms found no hit:\t['ENSG00000238851_x_st', 'ENSG00000238852_st', 'ENSG00000238853_st', 'ENSG00000238854_st', 'ENSG0000\n",
      "1000 input query terms found no hit:\t['MI0000288_st', 'MI0000289_st', 'MI0000289_x_st', 'MI0000290_st', 'MI0000291_st', 'MI0000292_st', '\n",
      "1000 input query terms found no hit:\t['MI0004652_st', 'MI0004653_st', 'MI0004658_st', 'MI0004659_st', 'MI0004660_s_st', 'MI0004660_x_st',\n",
      "1000 input query terms found no hit:\t['MI0015978_s_st', 'MI0015979_s_st', 'MI0015980_s_st', 'MI0015981_x_st', 'MI0015982_s_st', 'MI001598\n",
      "1000 input query terms found no hit:\t['MI0022620_st', 'MI0022621_st', 'MI0022622_st', 'MI0022623_st', 'MI0022624_st', 'MI0022625_st', 'MI\n",
      "1000 input query terms found no hit:\t['MIMAT0000425_st', 'MIMAT0000426_st', 'MIMAT0000427_st', 'MIMAT0000428_st', 'MIMAT0000429_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0001452_st', 'MIMAT0001453_st', 'MIMAT0001454_st', 'MIMAT0001456_st', 'MIMAT0001457_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0002506_st', 'MIMAT0002507_st', 'MIMAT0002508_st', 'MIMAT0002509_st', 'MIMAT0002510_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0003524_st', 'MIMAT0003525_st', 'MIMAT0003526_st', 'MIMAT0003527_st', 'MIMAT0003528_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0004589_st', 'MIMAT0004590_st', 'MIMAT0004591_st', 'MIMAT0004592_st', 'MIMAT0004593_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0005614_st', 'MIMAT0005615_st', 'MIMAT0005616_st', 'MIMAT0005617_st', 'MIMAT0005618_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0006648_st', 'MIMAT0006649_st', 'MIMAT0006650_st', 'MIMAT0006651_st', 'MIMAT0006652_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0007659_st', 'MIMAT0007660_st', 'MIMAT0007661_st', 'MIMAT0007662_st', 'MIMAT0007663_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0008682_st', 'MIMAT0008683_st', 'MIMAT0008684_st', 'MIMAT0008685_st', 'MIMAT0008686_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0009708_st', 'MIMAT0009709_st', 'MIMAT0009710_st', 'MIMAT0009711_st', 'MIMAT0009712_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0011614_st', 'MIMAT0011615_st', 'MIMAT0011616_st', 'MIMAT0011617_st', 'MIMAT0011618_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0012640_st', 'MIMAT0012641_st', 'MIMAT0012642_st', 'MIMAT0012643_st', 'MIMAT0012644_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0013684_st', 'MIMAT0013685_st', 'MIMAT0013686_st', 'MIMAT0013687_st', 'MIMAT0013688_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0014717_st', 'MIMAT0014718_st', 'MIMAT0014719_st', 'MIMAT0014720_st', 'MIMAT0014721_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0015742_st', 'MIMAT0015743_st', 'MIMAT0015744_st', 'MIMAT0015745_st', 'MIMAT0015746_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0016755_st', 'MIMAT0016756_st', 'MIMAT0016757_st', 'MIMAT0016758_st', 'MIMAT0016759_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0017761_st', 'MIMAT0017762_st', 'MIMAT0017763_st', 'MIMAT0017764_st', 'MIMAT0017765_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0018770_st', 'MIMAT0018771_st', 'MIMAT0018772_st', 'MIMAT0018773_st', 'MIMAT0018774_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0019778_st', 'MIMAT0019779_st', 'MIMAT0019780_st', 'MIMAT0019781_st', 'MIMAT0019782_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0020801_st', 'MIMAT0020802_st', 'MIMAT0020803_st', 'MIMAT0020804_st', 'MIMAT0020805_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0021817_st', 'MIMAT0021818_st', 'MIMAT0021819_st', 'MIMAT0021820_st', 'MIMAT0021821_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0022831_st', 'MIMAT0022832_st', 'MIMAT0022833_st', 'MIMAT0022834_st', 'MIMAT0022835_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0023836_st', 'MIMAT0023837_st', 'MIMAT0023838_st', 'MIMAT0023839_st', 'MIMAT0023840_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0024840_st', 'MIMAT0024841_st', 'MIMAT0024842_st', 'MIMAT0024843_st', 'MIMAT0024844_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0025853_st', 'MIMAT0025854_st', 'MIMAT0025855_st', 'MIMAT0025856_st', 'MIMAT0025857_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0026855_st', 'MIMAT0026856_st', 'MIMAT0026857_st', 'MIMAT0026858_st', 'MIMAT0026859_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0027859_st', 'MIMAT0027860_st', 'MIMAT0027861_st', 'MIMAT0027862_st', 'MIMAT0027863_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0028861_st', 'MIMAT0028862_st', 'MIMAT0028863_st', 'MIMAT0028864_st', 'MIMAT0028865_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0029861_st', 'MIMAT0029862_st', 'MIMAT0029863_st', 'MIMAT0029864_st', 'MIMAT0029865_st', 'MIM\n",
      "1000 input query terms found no hit:\t['MIMAT0030861_st', 'MIMAT0030862_st', 'MIMAT0030863_st', 'MIMAT0030864_st', 'MIMAT0030865_st', 'MIM\n",
      "222 input query terms found no hit:\t['MIMAT0031861_st', 'MIMAT0031862_st', 'MIMAT0031863_st', 'MIMAT0031864_st', 'SNORA11B_st', 'SNORA11\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 53 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Alzheimers Disease', the least common label is '1.0' with 53 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Alzheimers Disease' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'best_cohort' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[72], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m merged_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, \u001b[43mbest_cohort\u001b[49m\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m))\n",
      "\u001b[1;31mNameError\u001b[0m: name 'best_cohort' is not defined"
     ]
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, best_cohort+'.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If both age and gender features are available, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  5. Discussion and report\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
