{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"toc_visible":true,"machine_shape":"hm","gpuType":"T4"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"},"accelerator":"GPU"},"cells":[{"cell_type":"markdown","source":["# Gold standard curation: Preprocessing and single-step regression"],"metadata":{"collapsed":false,"id":"NB8dJaS7ZWQI"}},{"cell_type":"markdown","source":["In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n","\n","1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n","2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"],"metadata":{"collapsed":false,"id":"T8BiY3jEZWQL"}},{"cell_type":"markdown","source":["# 1. Basic setup"],"metadata":{"collapsed":false,"id":"6rY1UNA-ZWQM"}},{"cell_type":"code","execution_count":2,"outputs":[],"source":["import os\n","\n","# Set your preferred name\n","USER = \"Shuyi\"\n","# Set the data and output directories\n","DATA_ROOT = '/content/drive/MyDrive/AI4Science_Public/data'\n","OUTPUT_ROOT = '/content/drive/MyDrive/AI4Science_Public/output1'\n","TRAIT = \"Sjögren's Syndrome\"\n","\n","OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n","JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n","if not os.path.exists(OUTPUT_DIR):\n","    os.makedirs(OUTPUT_DIR, exist_ok=True)\n","\n","# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n","NORMALIZE_GENE = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:07.313098989Z","start_time":"2023-12-31T21:19:07.312718292Z"},"id":"CPen6psLZWQM","executionInfo":{"status":"ok","timestamp":1704801876419,"user_tz":-480,"elapsed":2436,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jMTdsU43vVg3","outputId":"71366407-279e-46f7-e340-5177d108379e","ExecuteTime":{"end_time":"2023-12-31T21:19:07.680659936Z","start_time":"2023-12-31T21:19:07.666959145Z"},"executionInfo":{"status":"ok","timestamp":1704801864456,"user_tz":-480,"elapsed":30877,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}],"source":["# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n","\n","import os\n","from google.colab import drive\n","\n","drive.mount('/content/drive', force_remount=True)\n","proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n","os.chdir(proj_dir)"]},{"cell_type":"code","source":["import sys\n","sys.path.append('/content/drive/MyDrive/Colab Notebooks')\n","!pip install mygene\n","!pip install sparse_lmm"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ucJm9PCJaIf8","executionInfo":{"status":"ok","timestamp":1704801891144,"user_tz":-480,"elapsed":11105,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"ebbd8ec6-52ed-48fd-ffcb-6665953fa56d"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting mygene\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Collecting biothings-client>=0.2.6 (from mygene)\n","  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n","Requirement already satisfied: requests>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from biothings-client>=0.2.6->mygene) (2.31.0)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2023.11.17)\n","Installing collected packages: biothings-client, mygene\n","Successfully installed biothings-client-0.3.1 mygene-3.2.2\n","Collecting sparse_lmm\n","  Downloading sparse_lmm-0.6-py3-none-any.whl (9.6 kB)\n","Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from sparse_lmm) (1.11.4)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from sparse_lmm) (1.23.5)\n","Installing collected packages: sparse_lmm\n","Successfully installed sparse_lmm-0.6\n"]}]},{"cell_type":"markdown","source":["# 2. Data preprocessing and selection"],"metadata":{"collapsed":false,"id":"uuyW6piSZWQP"}},{"cell_type":"markdown","source":["## 2.1. The TCGA Xena dataset"],"metadata":{"collapsed":false,"id":"gpsSPrUeZWQP"}},{"cell_type":"markdown","source":["In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."],"metadata":{"collapsed":false,"id":"xRKc-zchZWQQ"}},{"cell_type":"code","execution_count":4,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['.DS_Store',\n"," 'TCGA_Esophageal_Cancer_(ESCA)',\n"," 'TCGA_Pheochromocytoma_Paraganglioma_(PCPG)',\n"," 'TCGA_Bile_Duct_Cancer_(CHOL)',\n"," 'TCGA_Kidney_Chromophobe_(KICH)',\n"," 'TCGA_Thyroid_Cancer_(THCA)',\n"," 'TCGA_Sarcoma_(SARC)',\n"," 'TCGA_Adrenocortical_Cancer_(ACC)',\n"," 'TCGA_Thymoma_(THYM)',\n"," 'TCGA_Glioblastoma_(GBM)']"]},"metadata":{},"execution_count":4}],"source":["import os\n","from utils import *\n","\n","dataset = 'TCGA'\n","dataset_dir = os.path.join(DATA_ROOT, dataset)\n","os.listdir(dataset_dir)[:10]"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.770095084Z","start_time":"2023-12-31T21:19:09.102451553Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"YPDQJCUYZWQQ","executionInfo":{"status":"ok","timestamp":1704801901263,"user_tz":-480,"elapsed":4874,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"7a3c662b-65ed-42f1-d462-2ef01c753055"}},{"cell_type":"markdown","source":["If no match is found, jump directly to GEO in Part 2.2"],"metadata":{"collapsed":false,"id":"mTPvkvIcZWQR"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["trait_subdir = \"TCGA_Breast_Cancer_(BRCA)\"\n","cohort = 'Xena'\n","# All the cancer traits in Xena are binary\n","trait_type = 'binary'\n","# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n","is_available = True\n","\n","clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_BRCA_clinicalMatrix')\n","genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_HiSeqV2_PANCAN.gz')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.918243934Z","start_time":"2023-12-31T21:19:09.909021135Z"},"id":"_7r3vQ4DZWQR"}},{"cell_type":"code","source":["import pandas as pd\n","\n","clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n","genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n","age_col = gender_col = None"],"metadata":{"id":"MudwB-_iz7sc","ExecuteTime":{"end_time":"2023-12-31T21:19:13.613198409Z","start_time":"2023-12-31T21:19:10.409688003Z"}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"outputs":[],"source":["_, clinical_data_cols = check_rows_and_columns(clinical_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.617006167Z","start_time":"2023-12-31T21:19:13.614080653Z"},"id":"At6no_ODZWQR"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"['AJCC_Stage_nature2012',\n 'Age_at_Initial_Pathologic_Diagnosis_nature2012',\n 'CN_Clusters_nature2012',\n 'Converted_Stage_nature2012',\n 'Days_to_Date_of_Last_Contact_nature2012',\n 'Days_to_date_of_Death_nature2012',\n 'ER_Status_nature2012',\n 'Gender_nature2012',\n 'HER2_Final_Status_nature2012',\n 'Integrated_Clusters_no_exp__nature2012']"},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data_cols[:10]"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661538569Z","start_time":"2023-12-31T21:19:13.615946213Z"},"id":"IXOwP6DCZWQR","outputId":"2243b5b6-bbe1-4e3e-8ea8-7266ebe54d54"}},{"cell_type":"markdown","source":["Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-28T20:20:11.841042813Z","start_time":"2023-12-28T20:20:11.834458206Z"},"id":"i3SiZnBCZWQS"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['AJCC_Stage_nature2012', 'Age_at_Initial_Pathologic_Diagnosis_nature2012', 'CN_Clusters_nature2012', 'Converted_Stage_nature2012', 'Days_to_Date_of_Last_Contact_nature2012', 'Days_to_date_of_Death_nature2012', 'ER_Status_nature2012', 'Gender_nature2012', 'HER2_Final_Status_nature2012', 'Integrated_Clusters_no_exp__nature2012', 'Integrated_Clusters_unsup_exp__nature2012', 'Integrated_Clusters_with_PAM50__nature2012', 'Metastasis_Coded_nature2012', 'Metastasis_nature2012', 'Node_Coded_nature2012', 'Node_nature2012', 'OS_Time_nature2012', 'OS_event_nature2012', 'PAM50Call_RNAseq', 'PAM50_mRNA_nature2012', 'PR_Status_nature2012', 'RPPA_Clusters_nature2012', 'SigClust_Intrinsic_mRNA_nature2012', 'SigClust_Unsupervised_mRNA_nature2012', 'Survival_Data_Form_nature2012', 'Tumor_T1_Coded_nature2012', 'Tumor_nature2012', 'Vital_Status_nature2012', '_INTEGRATION', '_PANCAN_CNA_PANCAN_K8', '_PANCAN_Cluster_Cluster_PANCAN', '_PANCAN_DNAMethyl_BRCA', '_PANCAN_DNAMethyl_PANCAN', '_PANCAN_RPPA_PANCAN_K8', '_PANCAN_UNC_RNAseq_PANCAN_K16', '_PANCAN_miRNA_PANCAN', '_PANCAN_mirna_BRCA', '_PANCAN_mutation_PANCAN', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_surgery_locoregional_procedure', 'additional_surgery_metastatic_procedure', 'age_at_initial_pathologic_diagnosis', 'anatomic_neoplasm_subdivision', 'axillary_lymph_node_stage_method_type', 'axillary_lymph_node_stage_other_method_descriptive_text', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'breast_cancer_surgery_margin_status', 'breast_carcinoma_estrogen_receptor_status', 'breast_carcinoma_immunohistochemistry_er_pos_finding_scale', 'breast_carcinoma_immunohistochemistry_pos_cell_score', 'breast_carcinoma_immunohistochemistry_prgstrn_rcptr_ps_fndng_scl', 'breast_carcinoma_primary_surgical_procedure_name', 'breast_carcinoma_progesterone_receptor_status', 'breast_carcinoma_surgical_procedure_name', 'breast_neoplasm_other_surgical_procedure_descriptive_text', 'cytokeratin_immunohistochemistry_staining_method_mcrmtstss_ndctr', 'days_to_additional_surgery_locoregional_procedure', 'days_to_additional_surgery_metastatic_procedure', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_last_known_alive', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'disease_code', 'distant_metastasis_present_ind2', 'er_detection_method_text', 'er_level_cell_percentage_category', 'fluorescence_in_st_hybrdztn_dgnstc_prcdr_chrmsm_17_sgnl_rslt_rng', 'followup_case_report_form_submission_reason', 'form_completion_date', 'gender', 'her2_and_centromere_17_positive_finding_other_measuremnt_scl_txt', 'her2_erbb_method_calculation_method_text', 'her2_erbb_pos_finding_cell_percent_category', 'her2_erbb_pos_finding_fluorescence_n_st_hybrdztn_clcltn_mthd_txt', 'her2_immunohistochemistry_level_result', 'her2_neu_and_centromere_17_copy_number_analysis_npt_ttl_nmbr_cnt', 'her2_neu_breast_carcinoma_copy_analysis_input_total_number', 'her2_neu_chromosone_17_signal_ratio_value', 'her2_neu_metastatic_breast_carcinoma_copy_analysis_inpt_ttl_nmbr', 'histological_type', 'histological_type_other', 'history_of_neoadjuvant_treatment', 'hr2_n_nd_cntrmr_17_cpy_nmbr_mtsttc_brst_crcnm_nlyss_npt_ttl_nmbr', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'immunohistochemistry_positive_cell_score', 'informed_consent_verified', 'init_pathology_dx_method_other', 'initial_pathologic_diagnosis_method', 'initial_weight', 'is_ffpe', 'lab_proc_her2_neu_immunohistochemistry_receptor_status', 'lab_procedure_her2_neu_in_situ_hybrid_outcome_type', 'lost_follow_up', 'lymph_node_examined_count', 'margin_status', 'menopause_status', 'metastatic_breast_carcinm_ps_fndng_prgstrn_rcptr_thr_msr_scl_txt', 'metastatic_breast_carcinom_lb_prc_hr2_n_mmnhstchmstry_rcptr_stts', 'metastatic_breast_carcinoma_erbb2_immunohistochemistry_levl_rslt', 'metastatic_breast_carcinoma_estrogen_receptor_detection_mthd_txt', 'metastatic_breast_carcinoma_estrogen_receptor_status', 'metastatic_breast_carcinoma_estrogen_receptr_lvl_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_her2_erbb_method_calculatin_mthd_txt', 'metastatic_breast_carcinoma_her2_erbb_pos_findng_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_her2_neu_chromosone_17_signal_rat_vl', 'metastatic_breast_carcinoma_immunhstchmstry_r_pstv_fndng_scl_typ', 'metastatic_breast_carcinoma_immunohistochemistry_er_pos_cell_scr', 'metastatic_breast_carcinoma_immunohistochemistry_pr_pos_cell_scr', 'metastatic_breast_carcinoma_lab_proc_hr2_n_n_st_hybrdztn_tcm_typ', 'metastatic_breast_carcinoma_pos_finding_hr2_rbb2_thr_msr_scl_txt', 'metastatic_breast_carcinoma_progestern_rcptr_lvl_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_progesterone_receptor_dtctn_mthd_txt', 'metastatic_breast_carcinoma_progesterone_receptor_status', 'metastatic_site_at_diagnosis', 'metastatic_site_at_diagnosis_other', 'methylation_Clusters_nature2012', 'miRNA_Clusters_nature2012', 'mtsttc_brst_crcnm_flrscnc_n_st_hybrdztn_dgnstc_prc_cntrmr_17_sgn', 'mtsttc_brst_crcnm_hr2_rbb_ps_fndng_flrscnc_n_st_hybrdztn_clcltn', 'mtsttc_brst_crcnm_mmnhstchmstry_prgstrn_rcptr_pstv_fndng_scl_typ', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'number_of_lymphnodes_positive_by_he', 'number_of_lymphnodes_positive_by_ihc', 'oct_embedded', 'other_dx', 'pathologic_M', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'person_neoplasm_cancer_status', 'pgr_detection_method_text', 'pos_finding_her2_erbb2_other_measurement_scale_text', 'pos_finding_metastatic_brst_crcnm_strgn_rcptr_thr_msrmnt_scl_txt', 'pos_finding_progesterone_receptor_other_measurement_scale_text', 'positive_finding_estrogen_receptor_other_measurement_scale_text', 'postoperative_rx_tx', 'primary_lymph_node_presentation_assessment', 'progesterone_receptor_level_cell_percent_category', 'project_code', 'radiation_therapy', 'sample_type', 'sample_type_id', 'surgical_procedure_purpose_other_text', 'system_version', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'vial_number', 'vital_status', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_BRCA_RPPA_RBN', '_GENOMIC_ID_TCGA_BRCA_mutation', '_GENOMIC_ID_TCGA_BRCA_PDMRNAseq', '_GENOMIC_ID_TCGA_BRCA_hMethyl450', '_GENOMIC_ID_TCGA_BRCA_RPPA', '_GENOMIC_ID_TCGA_BRCA_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_BRCA_mutation_curated_wustl_gene', '_GENOMIC_ID_TCGA_BRCA_hMethyl27', '_GENOMIC_ID_TCGA_BRCA_PDMarrayCNV', '_GENOMIC_ID_TCGA_BRCA_miRNA_HiSeq', '_GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene', '_GENOMIC_ID_TCGA_BRCA_miRNA_GA', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile', '_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene', '_GENOMIC_ID_TCGA_BRCA_gistic2thd', '_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_BRCA_G4502A_07_3', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2', '_GENOMIC_ID_TCGA_BRCA_gistic2', '_GENOMIC_ID_TCGA_BRCA_PDMarray']\\n\""},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["f'''\n","Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n","candidate_age_cols = [col_name1, col_name2, ...]\n","candidate_gender_cols = [col_name1, col_name2, ...]\n","If no columns match a criterion, please provide an empty list.\n","\n","Column names:\n","{clinical_data_cols}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661858700Z","start_time":"2023-12-31T21:19:13.656749487Z"},"id":"x8P_531oZWQS","outputId":"eb07d863-f1f8-4417-e325-7734de5b48ab"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["candidate_age_cols = ['Age_at_Initial_Pathologic_Diagnosis_nature2012', 'age_at_initial_pathologic_diagnosis',\n","                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n","candidate_gender_cols = ['Gender_nature2012', 'gender']"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661938833Z","start_time":"2023-12-31T21:19:13.656901848Z"},"id":"2OWLgaibZWQS"}},{"cell_type":"markdown","source":["Choose a single column from the candidate columns that record age and gender information respectively.\n","If no column meets the requirement, keep 'age_col' or 'gender_col' to None"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:57:44.207565572Z","start_time":"2023-12-31T03:57:44.202177544Z"},"id":"o29AdGTRZWQS"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"{'Age_at_Initial_Pathologic_Diagnosis_nature2012': [nan, nan, nan, nan, nan],\n 'age_at_initial_pathologic_diagnosis': [55.0, 50.0, 62.0, 52.0, 50.0],\n 'days_to_birth': [-20211.0, -18538.0, -22848.0, -19074.0, -18371.0],\n 'year_of_initial_pathologic_diagnosis': [2004.0,\n  2003.0,\n  2011.0,\n  2011.0,\n  2013.0]}"},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_age_cols])"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.662078139Z","start_time":"2023-12-31T21:19:13.657006017Z"},"id":"ZES_DWC8ZWQS","outputId":"d8e4ce4f-a85a-406a-8b05-a36da9f5a39e"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["age_col = 'age_at_initial_pathologic_diagnosis'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.777159137Z","start_time":"2023-12-31T21:19:13.756210135Z"},"id":"Shy0HJpkZWQS"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"{'Gender_nature2012': [nan, nan, nan, nan, nan],\n 'gender': ['FEMALE', 'FEMALE', 'FEMALE', 'FEMALE', 'FEMALE']}"},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_gender_cols])"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.313425023Z","start_time":"2023-12-31T21:19:14.305221062Z"},"id":"2dEG2RPBZWQS","outputId":"1886b6fc-1943-4725-976e-349a76bf140c"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["gender_col = 'gender'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.852502962Z","start_time":"2023-12-31T21:19:14.845153121Z"},"id":"kNH4VpYKZWQT"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col, gender_col)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.246065025Z","start_time":"2023-12-31T21:19:15.240279162Z"},"id":"AGQ0X4scZWQT"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.593478652Z","start_time":"2023-12-31T21:19:15.583841656Z"},"id":"sxc-Bd_XZWQT"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"                 Breast Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17  \\\nsampleID                                                                      \nTCGA-3C-AAAU-01              1  55.0     0.0   0.607308 -3.194126 -0.531035   \nTCGA-3C-AALI-01              1  50.0     0.0  -0.641192 -4.928226  0.095465   \nTCGA-3C-AALJ-01              1  62.0     0.0   1.082808 -4.623726 -0.531035   \nTCGA-3C-AALK-01              1  52.0     0.0   0.121608 -2.881526 -0.531035   \nTCGA-4H-AAAK-01              1  50.0     0.0   0.420208 -3.282726 -0.531035   \n\n                    RNF10     RNF11    RNF13   GTF2IP1  ...     TULP2  \\\nsampleID                                                ...             \nTCGA-3C-AAAU-01 -0.145872  0.237422 -0.29921 -0.142694  ... -0.748878   \nTCGA-3C-AALI-01  0.098128 -0.541978 -0.32291 -0.044694  ...  0.312922   \nTCGA-3C-AALJ-01  0.484028 -0.183678 -0.91901  0.261106  ... -0.748878   \nTCGA-3C-AALK-01  0.179128  0.039222 -0.45491  0.180306  ...  0.415422   \nTCGA-4H-AAAK-01 -0.020972 -0.117978 -0.55781 -0.173794  ...  0.685222   \n\n                    NPY5R     GNGT2    GNGT1     TULP3      PTRF     BCL6B  \\\nsampleID                                                                     \nTCGA-3C-AAAU-01  0.565583 -0.767233 -1.28139 -0.271377 -0.492286  0.360373   \nTCGA-3C-AALI-01  0.079683  0.475267  1.99661  0.161423  0.273714  0.644673   \nTCGA-3C-AALJ-01 -0.656117 -0.216733  0.21081  0.101023  0.660514  1.295073   \nTCGA-3C-AALK-01 -0.178417 -0.211233 -1.28139  0.108023  1.132814  0.496773   \nTCGA-4H-AAAK-01 -0.698717 -0.484233 -1.28139 -0.223577  1.222714  0.225573   \n\n                    GSTK1      SELP      SELS  \nsampleID                                       \nTCGA-3C-AAAU-01  1.067905  0.076267 -0.392212  \nTCGA-3C-AALI-01 -0.428695  0.068667 -0.043812  \nTCGA-3C-AALJ-01  0.915105  0.168567  0.047788  \nTCGA-3C-AALK-01  0.240105  3.099767  0.112888  \nTCGA-4H-AAAK-01 -0.301995  0.200067 -0.080212  \n\n[5 rows x 20533 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Breast Cancer</th>\n      <th>Age</th>\n      <th>Gender</th>\n      <th>ARHGEF10L</th>\n      <th>HIF3A</th>\n      <th>RNF17</th>\n      <th>RNF10</th>\n      <th>RNF11</th>\n      <th>RNF13</th>\n      <th>GTF2IP1</th>\n      <th>...</th>\n      <th>TULP2</th>\n      <th>NPY5R</th>\n      <th>GNGT2</th>\n      <th>GNGT1</th>\n      <th>TULP3</th>\n      <th>PTRF</th>\n      <th>BCL6B</th>\n      <th>GSTK1</th>\n      <th>SELP</th>\n      <th>SELS</th>\n    </tr>\n    <tr>\n      <th>sampleID</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>TCGA-3C-AAAU-01</th>\n      <td>1</td>\n      <td>55.0</td>\n      <td>0.0</td>\n      <td>0.607308</td>\n      <td>-3.194126</td>\n      <td>-0.531035</td>\n      <td>-0.145872</td>\n      <td>0.237422</td>\n      <td>-0.29921</td>\n      <td>-0.142694</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>0.565583</td>\n      <td>-0.767233</td>\n      <td>-1.28139</td>\n      <td>-0.271377</td>\n      <td>-0.492286</td>\n      <td>0.360373</td>\n      <td>1.067905</td>\n      <td>0.076267</td>\n      <td>-0.392212</td>\n    </tr>\n    <tr>\n      <th>TCGA-3C-AALI-01</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>-0.641192</td>\n      <td>-4.928226</td>\n      <td>0.095465</td>\n      <td>0.098128</td>\n      <td>-0.541978</td>\n      <td>-0.32291</td>\n      <td>-0.044694</td>\n      <td>...</td>\n      <td>0.312922</td>\n      <td>0.079683</td>\n      <td>0.475267</td>\n      <td>1.99661</td>\n      <td>0.161423</td>\n      <td>0.273714</td>\n      <td>0.644673</td>\n      <td>-0.428695</td>\n      <td>0.068667</td>\n      <td>-0.043812</td>\n    </tr>\n    <tr>\n      <th>TCGA-3C-AALJ-01</th>\n      <td>1</td>\n      <td>62.0</td>\n      <td>0.0</td>\n      <td>1.082808</td>\n      <td>-4.623726</td>\n      <td>-0.531035</td>\n      <td>0.484028</td>\n      <td>-0.183678</td>\n      <td>-0.91901</td>\n      <td>0.261106</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>-0.656117</td>\n      <td>-0.216733</td>\n      <td>0.21081</td>\n      <td>0.101023</td>\n      <td>0.660514</td>\n      <td>1.295073</td>\n      <td>0.915105</td>\n      <td>0.168567</td>\n      <td>0.047788</td>\n    </tr>\n    <tr>\n      <th>TCGA-3C-AALK-01</th>\n      <td>1</td>\n      <td>52.0</td>\n      <td>0.0</td>\n      <td>0.121608</td>\n      <td>-2.881526</td>\n      <td>-0.531035</td>\n      <td>0.179128</td>\n      <td>0.039222</td>\n      <td>-0.45491</td>\n      <td>0.180306</td>\n      <td>...</td>\n      <td>0.415422</td>\n      <td>-0.178417</td>\n      <td>-0.211233</td>\n      <td>-1.28139</td>\n      <td>0.108023</td>\n      <td>1.132814</td>\n      <td>0.496773</td>\n      <td>0.240105</td>\n      <td>3.099767</td>\n      <td>0.112888</td>\n    </tr>\n    <tr>\n      <th>TCGA-4H-AAAK-01</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>0.420208</td>\n      <td>-3.282726</td>\n      <td>-0.531035</td>\n      <td>-0.020972</td>\n      <td>-0.117978</td>\n      <td>-0.55781</td>\n      <td>-0.173794</td>\n      <td>...</td>\n      <td>0.685222</td>\n      <td>-0.698717</td>\n      <td>-0.484233</td>\n      <td>-1.28139</td>\n      <td>-0.223577</td>\n      <td>1.222714</td>\n      <td>0.225573</td>\n      <td>-0.301995</td>\n      <td>0.200067</td>\n      <td>-0.080212</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 20533 columns</p>\n</div>"},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n","merged_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.074598194Z","start_time":"2023-12-31T21:19:15.973600157Z"},"id":"52Ks-OVpZWQT","outputId":"b72e5cfb-7d88-4ba3-f56e-34d28cec353f"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["The merged dataset contains 1215 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.341121723Z","start_time":"2023-12-31T21:19:16.330544222Z"},"id":"LQ-LqBo2ZWQT","outputId":"f075cebe-3783-4938-a706-5757fb540a7e"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["For the feature 'Breast Cancer', the least common label is '0' with 113 occurrences. This represents 9.30% of the dataset.\n","The distribution of the feature 'Breast Cancer' in this dataset is fine.\n","\n","Quartiles for 'Age':\n","  25%: 48.0\n","  50% (Median): 58.0\n","  75%: 67.0\n","Min: 26.0\n","Max: 90.0\n","The distribution of the feature 'Age' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '1.0' with 13 occurrences. This represents 1.07% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"data":{"text/plain":"False"},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:17.222505900Z","start_time":"2023-12-31T21:19:17.218781412Z"},"id":"aSrJydC2ZWQT","outputId":"8c4d2348-fb27-4df0-eba3-9a477b9e4038"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.401536463Z","start_time":"2023-12-31T21:19:17.980683628Z"},"id":"ldageITxZWQT"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.405288661Z","start_time":"2023-12-31T21:19:35.402799553Z"},"id":"Kq1zPqtfZWQT"}},{"cell_type":"markdown","source":["## 2.2. The GEO dataset"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-28T21:28:38.028916303Z","start_time":"2023-12-28T21:28:38.016245426Z"},"id":"zNMld4NxZWQT"}},{"cell_type":"markdown","source":["In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:25:23.253882615Z","start_time":"2023-12-31T03:25:23.244062710Z"},"id":"uGxJWgJIZWQT"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['GSE66795',\n"," 'GSE84844',\n"," 'GSE51092',\n"," 'GSE93683',\n"," 'GSE75679',\n"," 'GSE94510',\n"," 'GSE140161',\n"," 'GSE143153',\n"," 'GSE135809',\n"," 'GSE110007',\n"," 'GSE166373',\n"," 'GSE40611']"]},"metadata":{},"execution_count":5}],"source":["dataset = 'GEO'\n","trait_subdir = \"Sjögren's-Syndrome\"\n","\n","trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n","os.listdir(trait_path)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.411177575Z","start_time":"2023-12-31T21:19:35.405710636Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"8-VIPj3fZWQU","executionInfo":{"status":"ok","timestamp":1704457632099,"user_tz":-480,"elapsed":2870,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"2d35e0cc-3fc1-408f-abba-80a5ab29051a"}},{"cell_type":"markdown","source":["Repeat the below steps for all the accession numbers"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:39:42.634870142Z","start_time":"2023-12-31T03:39:42.534093295Z"},"id":"KJEJBf79ZWQU"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(\"/content/drive/MyDrive/AI4Science_Public/data/GEO/Sjögren's-Syndrome/GSE40611/GSE40611_family.soft.gz\",\n"," \"/content/drive/MyDrive/AI4Science_Public/data/GEO/Sjögren's-Syndrome/GSE40611/GSE40611_series_matrix.txt.gz\")"]},"metadata":{},"execution_count":51}],"source":["cohort = accession_num = \"GSE40611\"\n","cohort_dir = os.path.join(trait_path, accession_num)\n","soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n","soft_file, matrix_file"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.413836016Z","start_time":"2023-12-31T21:19:35.410443976Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"cCUVGG5KZWQU","executionInfo":{"status":"ok","timestamp":1704458488203,"user_tz":-480,"elapsed":528,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"5bdcbe32-f952-4f19-ee09-1d82fabd7302"}},{"cell_type":"markdown","source":["### Inital filtering and clinical data preprocessing"],"metadata":{"collapsed":false,"id":"Rqk_JLieZWQU"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["!Series_title\t\"Gene expression data of parotid tissue from Primary Sjogren’s Syndrome and controls\"\n","!Series_summary\t\"Primary Sjögren’s syndrome (pSS) is a chronic autoimmune disease with complex etiopathogenesis. Here we use Affymetrix U133 plus 2.0 microarray gene expression data from human parotid tissue. Parotid gland tissues were harvested from 17 pSS and 14 14 non-pSS sicca patients and 18 controls. The data were used in the following article:  Nazmul-Hossain ANM, Pollard RPE, Kroese FGM, Vissink A, Kallenberg CGM, Spijkervet FKL, Bootsma H, Michie SA, Gorr SU, Peck AB, Cai C, Zhou H, Horvath S, Wong DTW (2012) Systems Analysis of Primary Sjögren’s Syndrome Pathogenesis in Salivary Glands: Comparative Pathways and Molecular Events in Humans and a Mouse Model.\"\n","!Series_overall_design\t\"Parotid gland tissues were harvested from 17 pSS and 14 non-pSS sicca patients and 18 controls.\"\n"]}],"source":["background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n","clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n","\n","background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n","print(background_info)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.492119842Z","start_time":"2023-12-31T21:19:35.413383102Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"B0jX7hOMZWQU","executionInfo":{"status":"ok","timestamp":1704458492509,"user_tz":-480,"elapsed":936,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"485bab80-71e1-47dc-edeb-bd642ce285f6"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["         !Sample_geo_accession                GSM997850  \\\n","0  !Sample_characteristics_ch1  disease status: Control   \n","1  !Sample_characteristics_ch1                 batch: 1   \n","\n","                 GSM997851                GSM997852                GSM997853  \\\n","0  disease status: Control  disease status: Control  disease status: Control   \n","1                 batch: 1                 batch: 1                 batch: 1   \n","\n","                 GSM997854                GSM997855                GSM997856  \\\n","0  disease status: Control  disease status: Control  disease status: Control   \n","1                 batch: 1                 batch: 1                 batch: 1   \n","\n","                 GSM997857                GSM997858  ...  \\\n","0  disease status: Control  disease status: Control  ...   \n","1                 batch: 1                 batch: 1  ...   \n","\n","               GSM997889              GSM997890              GSM997891  \\\n","0  disease status: Sicca  disease status: Sicca  disease status: Sicca   \n","1               batch: 3               batch: 3               batch: 3   \n","\n","             GSM997892            GSM997893            GSM997894  \\\n","0  disease status: pSS  disease status: pSS  disease status: pSS   \n","1             batch: 2             batch: 2             batch: 2   \n","\n","             GSM997895            GSM997896            GSM997897  \\\n","0  disease status: pSS  disease status: pSS  disease status: pSS   \n","1             batch: 2             batch: 3             batch: 3   \n","\n","             GSM997898  \n","0  disease status: pSS  \n","1             batch: 3  \n","\n","[2 rows x 50 columns]"],"text/html":["\n","  <div id=\"df-11ea7a97-3c98-4f22-a019-1e65de458f3d\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>!Sample_geo_accession</th>\n","      <th>GSM997850</th>\n","      <th>GSM997851</th>\n","      <th>GSM997852</th>\n","      <th>GSM997853</th>\n","      <th>GSM997854</th>\n","      <th>GSM997855</th>\n","      <th>GSM997856</th>\n","      <th>GSM997857</th>\n","      <th>GSM997858</th>\n","      <th>...</th>\n","      <th>GSM997889</th>\n","      <th>GSM997890</th>\n","      <th>GSM997891</th>\n","      <th>GSM997892</th>\n","      <th>GSM997893</th>\n","      <th>GSM997894</th>\n","      <th>GSM997895</th>\n","      <th>GSM997896</th>\n","      <th>GSM997897</th>\n","      <th>GSM997898</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>disease status: Control</td>\n","      <td>...</td>\n","      <td>disease status: Sicca</td>\n","      <td>disease status: Sicca</td>\n","      <td>disease status: Sicca</td>\n","      <td>disease status: pSS</td>\n","      <td>disease status: pSS</td>\n","      <td>disease status: pSS</td>\n","      <td>disease status: pSS</td>\n","      <td>disease status: pSS</td>\n","      <td>disease status: pSS</td>\n","      <td>disease status: pSS</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>batch: 1</td>\n","      <td>...</td>\n","      <td>batch: 3</td>\n","      <td>batch: 3</td>\n","      <td>batch: 3</td>\n","      <td>batch: 2</td>\n","      <td>batch: 2</td>\n","      <td>batch: 2</td>\n","      <td>batch: 2</td>\n","      <td>batch: 3</td>\n","      <td>batch: 3</td>\n","      <td>batch: 3</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>2 rows × 50 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-11ea7a97-3c98-4f22-a019-1e65de458f3d')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-11ea7a97-3c98-4f22-a019-1e65de458f3d button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-11ea7a97-3c98-4f22-a019-1e65de458f3d');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-3245c9c0-ba80-4c2c-9361-13ec7e6dd23a\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-3245c9c0-ba80-4c2c-9361-13ec7e6dd23a')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-3245c9c0-ba80-4c2c-9361-13ec7e6dd23a button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":53}],"source":["clinical_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.507840422Z","start_time":"2023-12-31T21:19:35.451789107Z"},"colab":{"base_uri":"https://localhost:8080/","height":196},"id":"_oQdQ_4PZWQU","executionInfo":{"status":"ok","timestamp":1704458495763,"user_tz":-480,"elapsed":355,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"b3c81920-25a8-4820-d4ec-65276601e908"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{0: ['disease status: Control',\n","  'disease status: pSS',\n","  'disease status: Sicca'],\n"," 1: ['batch: 1', 'batch: 2', 'batch: 3']}"]},"metadata":{},"execution_count":54}],"source":["clinical_data_unique = get_unique_values_by_row(clinical_data)\n","clinical_data_unique"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508015426Z","start_time":"2023-12-31T21:19:35.492746637Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"jym8RXfxZWQU","executionInfo":{"status":"ok","timestamp":1704458498567,"user_tz":-480,"elapsed":402,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"b3bf7792-a6cc-4426-dab1-677bf021d602"}},{"cell_type":"markdown","source":["Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978204446Z","start_time":"2023-12-31T03:58:04.922270095Z"},"id":"a9Pe1BKHZWQU"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Sjögren\\'s Syndrome\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Sjögren\\'s Syndrome\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Gene expression data of parotid tissue from Primary Sjogren’s Syndrome and controls\"\\n!Series_summary\\t\"Primary Sjögren’s syndrome (pSS) is a chronic autoimmune disease with complex etiopathogenesis. Here we use Affymetrix U133 plus 2.0 microarray gene expression data from human parotid tissue. Parotid gland tissues were harvested from 17 pSS and 14 14 non-pSS sicca patients and 18 controls. The data were used in the following article:  Nazmul-Hossain ANM, Pollard RPE, Kroese FGM, Vissink A, Kallenberg CGM, Spijkervet FKL, Bootsma H, Michie SA, Gorr SU, Peck AB, Cai C, Zhou H, Horvath S, Wong DTW (2012) Systems Analysis of Primary Sjögren’s Syndrome Pathogenesis in Salivary Glands: Comparative Pathways and Molecular Events in Humans and a Mouse Model.\"\\n!Series_overall_design\\t\"Parotid gland tissues were harvested from 17 pSS and 14 non-pSS sicca patients and 18 controls.\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'disease status: Control\\', \\'disease status: pSS\\', \\'disease status: Sicca\\'], 1: [\\'batch: 1\\', \\'batch: 2\\', \\'batch: 3\\']}\\n'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":55}],"source":["f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n","1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n","2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n","   (1) Is there human data available for this trait?\n","   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n","   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n","   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n","\n","Background information about the dataset:\n","{background_info}\n","\n","Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n","{clinical_data_unique}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508232400Z","start_time":"2023-12-31T21:19:35.492903487Z"},"colab":{"base_uri":"https://localhost:8080/","height":178},"id":"VO8VaFm6ZWQa","executionInfo":{"status":"ok","timestamp":1704458500680,"user_tz":-480,"elapsed":351,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"b5a36011-b984-478c-e07c-a36b7bc927ea"}},{"cell_type":"markdown","source":["Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n","Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978240504Z","start_time":"2023-12-31T03:58:04.922365324Z"},"id":"00OpftSgZWQb"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["is_gene_availabe = True\n","trait_row_id = 0\n","age_row_id = None\n","gender_row_id = None\n","\n","trait_type = 'binary'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508295660Z","start_time":"2023-12-31T21:19:35.493015671Z"},"id":"boW_XbcAZWQb"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["is_available = is_gene_availabe and (trait_row_id is not None)\n","if not is_available:\n","    save_cohort_info(cohort, JSON_PATH, is_available)\n","    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508365153Z","start_time":"2023-12-31T21:19:35.493151450Z"},"id":"oHwwe4oPZWQb"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["# Verify and use the functions generated by GPT\n","\n","def convert_trait(trait_value):\n","    # For Sjögren's Syndrome (binary)\n","    if trait_value == 'disease status: Control':\n","        return 0\n","    elif trait_value == 'disease status: pSS' :\n","        return 1\n","    elif trait_value == 'disease status: Sicca':\n","        return 0\n","    else:\n","        return None\n","\n","def convert_age(age_value):\n","    # Assuming age is a continuous variable\n","    try:\n","        return float(age_value)\n","    except ValueError:\n","        return None\n","\n","def convert_gender(gender_value):\n","    # For gender (binary)\n","    if gender_value == 'gender: Female':\n","        return 1\n","    elif gender_value == 'gender: Female':\n","        return 0\n","    else:\n","        return None\n","\n","\n"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508426069Z","start_time":"2023-12-31T21:19:35.493220101Z"},"id":"uFVuqxemZWQb"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["                    GSM997850  GSM997851  GSM997852  GSM997853  GSM997854  \\\n","Sjögren's Syndrome          0          0          0          0          0   \n","\n","                    GSM997855  GSM997856  GSM997857  GSM997858  GSM997859  \\\n","Sjögren's Syndrome          0          0          0          0          0   \n","\n","                    ...  GSM997889  GSM997890  GSM997891  GSM997892  \\\n","Sjögren's Syndrome  ...          0          0          0          1   \n","\n","                    GSM997893  GSM997894  GSM997895  GSM997896  GSM997897  \\\n","Sjögren's Syndrome          1          1          1          1          1   \n","\n","                    GSM997898  \n","Sjögren's Syndrome          1  \n","\n","[1 rows x 49 columns]"],"text/html":["\n","  <div id=\"df-6dc6f782-59fb-4b0f-84ef-8e651cf69d4a\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GSM997850</th>\n","      <th>GSM997851</th>\n","      <th>GSM997852</th>\n","      <th>GSM997853</th>\n","      <th>GSM997854</th>\n","      <th>GSM997855</th>\n","      <th>GSM997856</th>\n","      <th>GSM997857</th>\n","      <th>GSM997858</th>\n","      <th>GSM997859</th>\n","      <th>...</th>\n","      <th>GSM997889</th>\n","      <th>GSM997890</th>\n","      <th>GSM997891</th>\n","      <th>GSM997892</th>\n","      <th>GSM997893</th>\n","      <th>GSM997894</th>\n","      <th>GSM997895</th>\n","      <th>GSM997896</th>\n","      <th>GSM997897</th>\n","      <th>GSM997898</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>Sjögren's Syndrome</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1 rows × 49 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6dc6f782-59fb-4b0f-84ef-8e651cf69d4a')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-6dc6f782-59fb-4b0f-84ef-8e651cf69d4a button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-6dc6f782-59fb-4b0f-84ef-8e651cf69d4a');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":60}],"source":["selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row_id, age_row_id, gender_row_id,\n","                                                      convert_trait, convert_age, convert_gender)\n","selected_clinical_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508533013Z","start_time":"2023-12-31T21:19:35.493272992Z"},"colab":{"base_uri":"https://localhost:8080/","height":147},"id":"3kQDZKKHZWQb","executionInfo":{"status":"ok","timestamp":1704458978224,"user_tz":-480,"elapsed":376,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"0f395400-b54e-4673-832b-5521472eae7d"}},{"cell_type":"markdown","source":["### Genetic data preprocessing and final filtering"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978664193Z","start_time":"2023-12-31T03:58:04.966117261Z"},"id":"FYVzuozPZWQb"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["          ID  GSM997850   GSM997851   GSM997852  GSM997853  GSM997854  \\\n","0  1007_s_at  474.82120  3764.07900  2193.57600  2712.1860  4876.2490   \n","1    1053_at   64.70827    78.76714   180.41580   424.7030   102.7654   \n","2     117_at  202.39090   159.54120   139.19140   625.5750   696.8760   \n","3     121_at  728.33920   253.09080   475.95720   539.3690   968.9207   \n","4  1255_g_at  309.24050   342.40070    25.24212   117.4153   446.7377   \n","\n","    GSM997855   GSM997856   GSM997857  GSM997858  ...   GSM997889   GSM997890  \\\n","0  4297.08400    81.72649  4146.96500  3418.5030  ...  4603.12700  2178.83300   \n","1   307.33720   400.26320   135.27490   137.9451  ...   449.11490    21.92653   \n","2   105.85680   390.57540    52.02566   477.9807  ...   259.49240   171.23530   \n","3   467.66870  2066.34200   783.54440  1392.0420  ...   739.03540   486.72830   \n","4    28.12048  1490.31800    92.12761   419.9179  ...    18.80592   110.53640   \n","\n","    GSM997891  GSM997892    GSM997893    GSM997894   GSM997895  GSM997896  \\\n","0  5903.17100  888.87700  1480.285000  3165.808000  3920.06600  995.23600   \n","1   366.58930  127.12930   265.340700   421.790700   510.07830   75.41043   \n","2   322.63720   19.56303   100.053800   299.769900   488.14100   42.98090   \n","3   657.47960  423.60690   364.608000   388.616100   507.78380  367.69270   \n","4    20.58882   18.15357     8.098538     6.897666    82.34903   46.26534   \n","\n","    GSM997897   GSM997898  \n","0  1269.23500  2221.01000  \n","1   164.05850   155.66470  \n","2    43.12577    67.33868  \n","3   391.64960   383.42200  \n","4    52.66693    32.73994  \n","\n","[5 rows x 50 columns]"],"text/html":["\n","  <div id=\"df-39e0ae1b-8862-4143-ae72-f85cd9261b92\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ID</th>\n","      <th>GSM997850</th>\n","      <th>GSM997851</th>\n","      <th>GSM997852</th>\n","      <th>GSM997853</th>\n","      <th>GSM997854</th>\n","      <th>GSM997855</th>\n","      <th>GSM997856</th>\n","      <th>GSM997857</th>\n","      <th>GSM997858</th>\n","      <th>...</th>\n","      <th>GSM997889</th>\n","      <th>GSM997890</th>\n","      <th>GSM997891</th>\n","      <th>GSM997892</th>\n","      <th>GSM997893</th>\n","      <th>GSM997894</th>\n","      <th>GSM997895</th>\n","      <th>GSM997896</th>\n","      <th>GSM997897</th>\n","      <th>GSM997898</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>1007_s_at</td>\n","      <td>474.82120</td>\n","      <td>3764.07900</td>\n","      <td>2193.57600</td>\n","      <td>2712.1860</td>\n","      <td>4876.2490</td>\n","      <td>4297.08400</td>\n","      <td>81.72649</td>\n","      <td>4146.96500</td>\n","      <td>3418.5030</td>\n","      <td>...</td>\n","      <td>4603.12700</td>\n","      <td>2178.83300</td>\n","      <td>5903.17100</td>\n","      <td>888.87700</td>\n","      <td>1480.285000</td>\n","      <td>3165.808000</td>\n","      <td>3920.06600</td>\n","      <td>995.23600</td>\n","      <td>1269.23500</td>\n","      <td>2221.01000</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1053_at</td>\n","      <td>64.70827</td>\n","      <td>78.76714</td>\n","      <td>180.41580</td>\n","      <td>424.7030</td>\n","      <td>102.7654</td>\n","      <td>307.33720</td>\n","      <td>400.26320</td>\n","      <td>135.27490</td>\n","      <td>137.9451</td>\n","      <td>...</td>\n","      <td>449.11490</td>\n","      <td>21.92653</td>\n","      <td>366.58930</td>\n","      <td>127.12930</td>\n","      <td>265.340700</td>\n","      <td>421.790700</td>\n","      <td>510.07830</td>\n","      <td>75.41043</td>\n","      <td>164.05850</td>\n","      <td>155.66470</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>117_at</td>\n","      <td>202.39090</td>\n","      <td>159.54120</td>\n","      <td>139.19140</td>\n","      <td>625.5750</td>\n","      <td>696.8760</td>\n","      <td>105.85680</td>\n","      <td>390.57540</td>\n","      <td>52.02566</td>\n","      <td>477.9807</td>\n","      <td>...</td>\n","      <td>259.49240</td>\n","      <td>171.23530</td>\n","      <td>322.63720</td>\n","      <td>19.56303</td>\n","      <td>100.053800</td>\n","      <td>299.769900</td>\n","      <td>488.14100</td>\n","      <td>42.98090</td>\n","      <td>43.12577</td>\n","      <td>67.33868</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>121_at</td>\n","      <td>728.33920</td>\n","      <td>253.09080</td>\n","      <td>475.95720</td>\n","      <td>539.3690</td>\n","      <td>968.9207</td>\n","      <td>467.66870</td>\n","      <td>2066.34200</td>\n","      <td>783.54440</td>\n","      <td>1392.0420</td>\n","      <td>...</td>\n","      <td>739.03540</td>\n","      <td>486.72830</td>\n","      <td>657.47960</td>\n","      <td>423.60690</td>\n","      <td>364.608000</td>\n","      <td>388.616100</td>\n","      <td>507.78380</td>\n","      <td>367.69270</td>\n","      <td>391.64960</td>\n","      <td>383.42200</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>1255_g_at</td>\n","      <td>309.24050</td>\n","      <td>342.40070</td>\n","      <td>25.24212</td>\n","      <td>117.4153</td>\n","      <td>446.7377</td>\n","      <td>28.12048</td>\n","      <td>1490.31800</td>\n","      <td>92.12761</td>\n","      <td>419.9179</td>\n","      <td>...</td>\n","      <td>18.80592</td>\n","      <td>110.53640</td>\n","      <td>20.58882</td>\n","      <td>18.15357</td>\n","      <td>8.098538</td>\n","      <td>6.897666</td>\n","      <td>82.34903</td>\n","      <td>46.26534</td>\n","      <td>52.66693</td>\n","      <td>32.73994</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 50 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-39e0ae1b-8862-4143-ae72-f85cd9261b92')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-39e0ae1b-8862-4143-ae72-f85cd9261b92 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-39e0ae1b-8862-4143-ae72-f85cd9261b92');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-8ee32df4-861f-4629-b40c-a2f102913977\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8ee32df4-861f-4629-b40c-a2f102913977')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-8ee32df4-861f-4629-b40c-a2f102913977 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":61}],"source":["genetic_data = get_genetic_data(matrix_file)\n","genetic_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590261560Z","start_time":"2023-12-31T21:19:35.493320362Z"},"colab":{"base_uri":"https://localhost:8080/","height":256},"id":"gVXr4nRrZWQb","executionInfo":{"status":"ok","timestamp":1704458999252,"user_tz":-480,"elapsed":818,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"7fd63494-056d-4d5e-e405-0376eaac4a24"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['1007_s_at',\n"," '1053_at',\n"," '117_at',\n"," '121_at',\n"," '1255_g_at',\n"," '1294_at',\n"," '1316_at',\n"," '1320_at',\n"," '1405_i_at',\n"," '1431_at',\n"," '1438_at',\n"," '1487_at',\n"," '1494_f_at',\n"," '1552256_a_at',\n"," '1552257_a_at',\n"," '1552258_at',\n"," '1552261_at',\n"," '1552263_at',\n"," '1552264_a_at',\n"," '1552266_at']"]},"metadata":{},"execution_count":62}],"source":["gene_row_ids = genetic_data['ID'][:20].tolist()\n","gene_row_ids"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590443507Z","start_time":"2023-12-31T21:19:35.536713559Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"j3WTypxBZWQc","executionInfo":{"status":"ok","timestamp":1704459002630,"user_tz":-480,"elapsed":396,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"d1ce5599-66e0-4935-ce6a-7de5c33807b7"}},{"cell_type":"markdown","source":["Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:30:41.595335164Z","start_time":"2023-12-31T03:30:41.513232329Z"},"id":"FrSj5umZZWQc"}},{"cell_type":"markdown","source":["Reference prompt:"],"metadata":{"collapsed":false,"id":"DpeO_Jd3ZWQc"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\\n\""],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":63}],"source":["f'''\n","Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n","requires_gene_mapping = (True or False)\n","\n","Row headers:\n","{gene_row_ids}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590590427Z","start_time":"2023-12-31T21:19:35.536780517Z"},"colab":{"base_uri":"https://localhost:8080/","height":125},"id":"kInJXlliZWQc","executionInfo":{"status":"ok","timestamp":1704459004766,"user_tz":-480,"elapsed":335,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"9042cc51-98b4-48e5-deff-dddb139a8b19"}},{"cell_type":"markdown","source":["If not required, jump directly to the gene normalization step"],"metadata":{"collapsed":false,"id":"wEp4DKmcZWQc"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["requires_gene_mapping = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590648558Z","start_time":"2023-12-31T21:19:35.536821375Z"},"id":"kv3cD6jRZWQc"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["if requires_gene_mapping:\n","    gene_annotation = get_gene_annotation(soft_file)[0]\n","    gene_annotation_summary = preview_df(gene_annotation)\n","    gene_annotation_summary"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590695457Z","start_time":"2023-12-31T21:19:35.536910836Z"},"id":"TBTqW54jZWQc"}},{"cell_type":"markdown","source":["Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978931927Z","start_time":"2023-12-31T03:58:04.966328339Z"},"id":"Lg3BW_bsZWQc"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n","    \n"]}],"source":["if requires_gene_mapping:\n","    print(f'''\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {gene_annotation_summary}\n","    ''')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590742236Z","start_time":"2023-12-31T21:19:35.537050893Z"},"id":"kY03dnyLZWQc","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459026324,"user_tz":-480,"elapsed":369,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"c9b83016-3af5-4eea-945a-376cbffd8080"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["if requires_gene_mapping:\n","    probe_id_key = 'ID'\n","    gene_symb_key = 'Gene Symbol'\n","    gene_mapping = get_gene_mapping(gene_annotation, probe_id_key, gene_symb_key)\n","    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590786881Z","start_time":"2023-12-31T21:19:35.537132859Z"},"id":"fD2yNE8vZWQd"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:23 input query terms found dup hits:\t[('A2MP1', 2), ('AACSP1', 2), ('AADACP1', 2), ('ABCC13', 2), ('ABCC6P1', 2), ('ACTA2-AS1', 2), ('ACT\n","WARNING:biothings.client:313 input query terms found no hit:\t['AAED1', 'AARS', 'AARSD1 /// PTGES3L /// PTGES3L-AARSD1', 'AB074162 /// MIR181A2HG', 'AB488780', 'A\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P2', 2), ('ANXA2P3', 2), ('AOC4P', 2), ('AOX2P', 2), ('AQP4-AS1', 2), ('ARHG\n","WARNING:biothings.client:321 input query terms found no hit:\t['ANXA2 /// NHP2L1', 'ANXA8 /// ANXA8L1', 'ANXA8 /// ANXA8L1 /// LOC100996760 /// LOC102725207', 'AP\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('BCORP1', 2), ('BCRP3', 2), ('BMS1P20', 2), ('BPIFA4P', 2), ('BRD7P3', 2), ('BTF3P11', 2), ('BTN2A\n","WARNING:biothings.client:385 input query terms found no hit:\t['BCMO1', 'BD495725', 'BGLAP /// PMF1 /// PMF1-BGLAP', 'BHLHB9', 'BIRC5 /// EPR-1', 'BLOC1S1-RDH5 //\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('CASC6', 2), ('CCDC144CP', 2), ('CCNT2-AS1', 2), ('CCNYL2', 2), ('CD99P1', 3), ('CDC42-IT1', 2), (\n","WARNING:biothings.client:132 input query terms found no hit:\t['CASC3 /// MIR6866', 'CASC4', 'CASC5', 'CASC7', 'CASP16', 'CATSPER2 /// CATSPER2P1 /// LOC101930343\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:21 input query terms found dup hits:\t[('CLRN1-AS1', 2), ('CLUHP3', 2), ('CLYBL-AS1', 2), ('CMAHP', 2), ('CNTNAP3P2', 2), ('COPG2IT1', 2),\n","WARNING:biothings.client:211 input query terms found no hit:\t['CLTC-IT1', 'CLUHP3 /// ZNF720', 'CMC4 /// MTCP1', 'CMIP /// LOC100129617', 'CNN2P10 /// CNN2P10', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:28 input query terms found dup hits:\t[('DDX11L2', 2), ('DEFB122', 2), ('DIRC1', 2), ('DIRC3', 2), ('DLEU2L', 2), ('DLG5-AS1', 3), ('DLGAP\n","WARNING:biothings.client:143 input query terms found no hit:\t['DDI2 /// RSC1A1', 'DDR1 /// MIR4640', 'DDR1-AS1', 'DDT /// DDTL', 'DDX11 /// DDX12P /// LOC642846'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:13 input query terms found dup hits:\t[('EPB41L4A-AS1', 2), ('ERVH-1', 2), ('ERVK13-1', 2), ('FAHD2CP', 2), ('FAM201A', 2), ('FAM224A', 2)\n","WARNING:biothings.client:274 input query terms found no hit:\t['EPB41L4A-AS2', 'EPN2 /// EPN2-IT1', 'EPOR /// RGL3', 'EPPIN /// EPPIN-WFDC6', 'EPRS', 'EPS15P1 ///\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('FOXP1-IT1', 2), ('FRMD8P1', 2), ('FRY-AS1', 2), ('FTH1P5', 2), ('FUT8-AS1', 2), ('GABARAPL3', 2),\n","WARNING:biothings.client:136 input query terms found no hit:\t['FOPNL', 'FOXD4 /// FOXD4L1', 'FOXO3 /// FOXO3B', 'FPGT-TNNI3K /// TNNI3K', 'FRG1 /// FRG1B /// LOC\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:35 input query terms found dup hits:\t[('GUSBP1', 2), ('GUSBP2', 2), ('HAGLROS', 2), ('HAR1A', 2), ('HBBP1', 2), ('HBD', 2), ('HCG4', 9), \n","WARNING:biothings.client:240 input query terms found no hit:\t['GUCY1B3', 'GUSBP1 /// GUSBP4', 'GUSBP1 /// GUSBP4 /// LOC100996497', 'GUSBP3 /// GUSBP9', 'GUSBP3 \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('INTS6-AS1', 2), ('IPO5P1', 2), ('IQCJ-SCHIP1-AS1', 2), ('ISM1-AS1', 2), ('JARID2-AS1', 2), ('JAZF\n","WARNING:biothings.client:159 input query terms found no hit:\t['IL4I1 /// NUP62', 'ILF3-AS1', 'IMPAD1', 'INADL', 'INO80B /// INO80B-WBP1', 'INO80B /// INO80B-WBP1\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:112 input query terms found dup hits:\t[('LBX1-AS1', 2), ('LIMD1-AS1', 2), ('LINC00029', 2), ('LINC00032', 2), ('LINC00112', 2), ('LINC0011\n","WARNING:biothings.client:326 input query terms found no hit:\t['LARS', 'LDHAL6CP /// LDHAL6CP', 'LDOC1L', 'LECT1', 'LEFTY1 /// LEFTY2', 'LEPR /// LEPROT', 'LEPRE1\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:769 input query terms found no hit:\t['LOC100505784', 'LOC100505797', 'LOC100505811', 'LOC100505812', 'LOC100505817', 'LOC100505824', 'LO\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:9 input query terms found dup hits:\t[('LPAL2', 2), ('LRP5L', 2), ('LRRC37A4P', 3), ('LRRC37A5P', 2), ('LRRC37BP1', 2), ('LY6G6E', 8), ('\n","WARNING:biothings.client:506 input query terms found no hit:\t['LOC102723661', 'LOC102723678 /// LOC102723709', 'LOC102723697', 'LOC102723721', 'LOC102723722 /// \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('MAPT-AS1', 4), ('MCF2L-AS1', 2), ('MED14OS', 2), ('MED15P9', 3), ('MEIS3P1', 2), ('MIATNB', 2), (\n","WARNING:biothings.client:183 input query terms found no hit:\t['MARC1', 'MARC2', 'MARCH1', 'MARCH10', 'MARCH11', 'MARCH2', 'MARCH3', 'MARCH4', 'MARCH5', 'MARCH6',\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:20 input query terms found dup hits:\t[('NAPA-AS1', 2), ('NAPSB', 2), ('NAT8B', 3), ('NAV2-AS2', 2), ('NAV2-AS4', 2), ('NAV2-IT1', 2), ('N\n","WARNING:biothings.client:86 input query terms found no hit:\t['MYLPF', 'MYNN /// RP11-362K14.5', 'MYO18A /// TIAF1', 'MZT2A /// MZT2B /// PHGDH', 'N4BP2L2 /// U5\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:28 input query terms found dup hits:\t[('OR4C1P', 2), ('OR52K3P', 2), ('OR5E1P', 4), ('OR6W1P', 3), ('OR7E104P', 2), ('OR7E12P', 2), ('OR7\n","WARNING:biothings.client:69 input query terms found no hit:\t['OR2N1P /// OR2N1P', 'OR4A1P /// OR4A1P', 'OR4F16 /// OR4F21 /// OR4F29 /// OR4F3', 'OR51A10P /// O\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:26 input query terms found dup hits:\t[('POM121L10P', 2), ('POM121L8P', 2), ('POM121L9P', 2), ('POTEKP', 2), ('POU6F2-AS2', 2), ('PPBPP2',\n","WARNING:biothings.client:100 input query terms found no hit:\t['PNMAL1', 'PNMAL2', 'POLDIP3 /// RRP7B', 'POLR2J /// POLR2J2 /// POLR2J3 /// POLR2J4 /// UPK3BL', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:11 input query terms found dup hits:\t[('RASSF8-AS1', 2), ('RBAKDN', 2), ('RBFADN', 2), ('RBMS3-AS3', 2), ('RBMY3AP', 3), ('REXO1L1P', 2),\n","WARNING:biothings.client:533 input query terms found no hit:\t['RBBP4P4 /// RBBP4P4', 'RBM14-RBM4 /// RBM4', 'RBMY1A1 /// RBMY1B /// RBMY1D /// RBMY1E /// RBMY1F \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:20 input query terms found dup hits:\t[('RP9P', 2), ('RPL23AP32', 2), ('RPL23AP53', 2), ('RPL32P3', 2), ('RPPH1', 3), ('RPS10P7', 2), ('RP\n","WARNING:biothings.client:387 input query terms found no hit:\t['RP11-560G2.2', 'RP11-567C2.1', 'RP11-567I13.1', 'RP11-568N6.1', 'RP11-573N10.1', 'RP11-574H6.1', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('SH3GL1P1', 2), ('SH3GL1P2', 2), ('SH3RF3-AS1', 2), ('SIGLEC17P', 2), ('SIGLEC5', 2), ('SLC26A4-AS\n","WARNING:biothings.client:70 input query terms found no hit:\t['SGK223', 'SGK494', 'SGOL1', 'SGOL2', 'SH3GL1P1 /// SH3GL1P2', 'SHANK2-AS2 /// SHANK2-AS2', 'SHC1P1\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('SPRR2C', 2), ('ST20', 2), ('STAG3L3', 2), ('STAG3L4', 2), ('STARD13-AS', 2), ('STARD7-AS1', 2), (\n","WARNING:biothings.client:88 input query terms found no hit:\t['SPTY2D1-AS1', 'SQRDL', 'SRCRB4D', 'SRGAP2 /// SRGAP2B /// SRGAP2C', 'SRGAP2B /// SRGAP2C /// SRGAP\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:33 input query terms found dup hits:\t[('TLR8-AS1', 2), ('TMCO5B', 2), ('TMEM108-AS1', 2), ('TMEM191A', 2), ('TMEM254-AS1', 2), ('TMEM44-A\n","WARNING:biothings.client:98 input query terms found no hit:\t['TMCC1-AS1', 'TMED7 /// TMED7-TICAM2', 'TMEM110', 'TMEM110 /// TMEM110-MUSTN1', 'TMEM133', 'TMEM136\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('UBXN7-AS1', 2), ('USP2-AS1', 2), ('USP3-AS1', 2), ('USP30-AS1', 2), ('USP32P2', 2), ('VN1R10P', 2\n","WARNING:biothings.client:89 input query terms found no hit:\t['UBBP1 /// UBBP1 /// UBBP4 /// UBBP4', 'UBBP2 /// UBBP2', 'UBE2E2-AS1', 'UBE2Q2L', 'UBE2V1P9 /// UB\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-520...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('ZNF213-AS1', 2), ('ZNF252P', 2), ('ZNF29P', 2), ('ZNF300P1', 2), ('ZNF542P', 2), ('ZNF663P', 2), \n","WARNING:biothings.client:32 input query terms found no hit:\t['ZNF208 /// ZNF595 /// ZNF718', 'ZNF259P1', 'ZNF259P1 /// ZPR1', 'ZNF271', 'ZNF285 /// ZNF806', 'ZN\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"]}],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590837318Z","start_time":"2023-12-31T21:19:35.537209586Z"},"id":"IYTbjGQJZWQd","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459095849,"user_tz":-480,"elapsed":45133,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"292b71c4-f3f8-4a09-e96f-9a9aa6a29f30"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n","# The preprocessing runs through, which means is_available should be True\n","is_available = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590902191Z","start_time":"2023-12-31T21:19:35.537285901Z"},"id":"gGEX1XUUZWQd"}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["The merged dataset contains 49 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.591095170Z","start_time":"2023-12-31T21:19:35.538708445Z"},"id":"Iy9Qpg0MZWQd","outputId":"0d0eb87c-0cfc-493f-c828-58e0c24aa0c3","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459104251,"user_tz":-480,"elapsed":344,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["For the feature 'Sjögren's Syndrome', the least common label is '1.0' with 17 occurrences. This represents 34.69% of the dataset.\n","The distribution of the feature 'Sjögren's Syndrome' in this dataset is fine.\n","\n"]},{"output_type":"execute_result","data":{"text/plain":["False"]},"metadata":{},"execution_count":71}],"source":["is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610849182Z","start_time":"2023-12-31T21:19:35.541210258Z"},"id":"LXD45pSgZWQd","outputId":"d3bc9d85-ea90-4a3f-f01c-9409b4c56e32","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459106373,"user_tz":-480,"elapsed":369,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610923234Z","start_time":"2023-12-31T21:19:35.584848473Z"},"id":"OZ0SAESiZWQd"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610970734Z","start_time":"2023-12-31T21:19:35.585042694Z"},"id":"02hAGq1wZWQd"}},{"cell_type":"markdown","source":["### 3. Do regression & Cross Validation"],"metadata":{"id":"-MTPhRGxJV7I","ExecuteTime":{"end_time":"2023-12-31T03:58:22.271683755Z","start_time":"2023-12-31T03:58:22.246557674Z"}}},{"cell_type":"code","execution_count":5,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'GSE84844'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":5}],"source":["best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH, 'has_age')\n","best_cohort"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.459442087Z","start_time":"2023-12-31T21:19:44.416219648Z"},"id":"2P5jb3pNZWQd","outputId":"39e069c3-7ee1-4bf8-9de1-0977d8061a7e","colab":{"base_uri":"https://localhost:8080/","height":35},"executionInfo":{"status":"ok","timestamp":1704801911498,"user_tz":-480,"elapsed":1053,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":6,"outputs":[{"output_type":"execute_result","data":{"text/plain":["   cohort_id  is_usable  is_available is_biased has_age has_gender  \\\n","1   GSE84844       True          True     False    True      False   \n","6  GSE143153       True          True     False    True      False   \n","8  GSE110007       True          True     False    True      False   \n","\n","   sample_size note  \n","1         60.0       \n","6         32.0       \n","8         29.0       "],"text/html":["\n","  <div id=\"df-ccf8dad8-fc16-43e6-8f02-a720f792a7d0\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>cohort_id</th>\n","      <th>is_usable</th>\n","      <th>is_available</th>\n","      <th>is_biased</th>\n","      <th>has_age</th>\n","      <th>has_gender</th>\n","      <th>sample_size</th>\n","      <th>note</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>1</th>\n","      <td>GSE84844</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>60.0</td>\n","      <td></td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>GSE143153</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>32.0</td>\n","      <td></td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>GSE110007</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>29.0</td>\n","      <td></td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ccf8dad8-fc16-43e6-8f02-a720f792a7d0')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-ccf8dad8-fc16-43e6-8f02-a720f792a7d0 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-ccf8dad8-fc16-43e6-8f02-a720f792a7d0');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-bc268d68-1145-4e0c-8020-7094eb71888e\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-bc268d68-1145-4e0c-8020-7094eb71888e')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-bc268d68-1145-4e0c-8020-7094eb71888e button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":6}],"source":["ranked_df.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.808945960Z","start_time":"2023-12-31T21:19:44.799467072Z"},"id":"ooo1W8o8ZWQd","outputId":"8dd0915a-7d3d-44a8-9cb5-319741fdcef8","colab":{"base_uri":"https://localhost:8080/","height":143},"executionInfo":{"status":"ok","timestamp":1704801914130,"user_tz":-480,"elapsed":5,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":7,"outputs":[],"source":["merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, best_cohort+'.csv'))"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.222719978Z","start_time":"2023-12-31T21:19:45.162873006Z"},"id":"4_XKlkP_ZWQd","executionInfo":{"status":"ok","timestamp":1704801919918,"user_tz":-480,"elapsed":1954,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":8,"outputs":[{"output_type":"execute_result","data":{"text/plain":["   Sjögren's Syndrome  Age      A1BG  A1BG-AS1      A1CF       A2M   A2M-AS1  \\\n","0                   0   45  5.739716  5.531918  4.073243  4.680171  8.261442   \n","1                   0   40  5.452101  5.131475  4.106471  5.052298  9.217795   \n","2                   0   54  5.784453  5.785170  3.978284  4.601635  8.902882   \n","3                   0   50  5.283689  5.385669  4.178392  4.617995  8.780384   \n","4                   0   31  5.414403  5.828662  3.836844  4.309026  8.221119   \n","\n","      A2ML1     A2MP1    A4GALT  ...    ZWILCH     ZWINT      ZXDA      ZXDB  \\\n","0  3.956358  5.001172  6.031274  ...  5.349392  6.811687  4.856520  5.001800   \n","1  4.260219  4.639341  6.057871  ...  5.193747  5.517230  5.310642  4.927082   \n","2  4.107508  4.809921  5.912525  ...  5.696565  5.896517  5.604909  5.107001   \n","3  3.917632  5.144286  6.223132  ...  5.053215  5.802501  5.288305  4.732636   \n","4  4.063579  4.561822  6.135777  ...  5.269230  6.576080  5.135130  4.916972   \n","\n","       ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n","0  7.012791  4.613131  7.855182  7.594642  6.093332  7.106518  \n","1  7.025551  4.992692  8.006711  7.850989  6.596223  6.962768  \n","2  6.872938  4.533009  8.112374  7.337760  6.405903  7.233815  \n","3  7.052588  4.810001  8.001139  7.901643  6.893890  6.624012  \n","4  6.706851  4.590855  7.819838  8.041268  6.623575  7.204644  \n","\n","[5 rows x 17872 columns]"],"text/html":["\n","  <div id=\"df-0e9ccda8-0771-4f98-b251-faa0ee942ba0\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Sjögren's Syndrome</th>\n","      <th>Age</th>\n","      <th>A1BG</th>\n","      <th>A1BG-AS1</th>\n","      <th>A1CF</th>\n","      <th>A2M</th>\n","      <th>A2M-AS1</th>\n","      <th>A2ML1</th>\n","      <th>A2MP1</th>\n","      <th>A4GALT</th>\n","      <th>...</th>\n","      <th>ZWILCH</th>\n","      <th>ZWINT</th>\n","      <th>ZXDA</th>\n","      <th>ZXDB</th>\n","      <th>ZXDC</th>\n","      <th>ZYG11A</th>\n","      <th>ZYG11B</th>\n","      <th>ZYX</th>\n","      <th>ZZEF1</th>\n","      <th>ZZZ3</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0</td>\n","      <td>45</td>\n","      <td>5.739716</td>\n","      <td>5.531918</td>\n","      <td>4.073243</td>\n","      <td>4.680171</td>\n","      <td>8.261442</td>\n","      <td>3.956358</td>\n","      <td>5.001172</td>\n","      <td>6.031274</td>\n","      <td>...</td>\n","      <td>5.349392</td>\n","      <td>6.811687</td>\n","      <td>4.856520</td>\n","      <td>5.001800</td>\n","      <td>7.012791</td>\n","      <td>4.613131</td>\n","      <td>7.855182</td>\n","      <td>7.594642</td>\n","      <td>6.093332</td>\n","      <td>7.106518</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0</td>\n","      <td>40</td>\n","      <td>5.452101</td>\n","      <td>5.131475</td>\n","      <td>4.106471</td>\n","      <td>5.052298</td>\n","      <td>9.217795</td>\n","      <td>4.260219</td>\n","      <td>4.639341</td>\n","      <td>6.057871</td>\n","      <td>...</td>\n","      <td>5.193747</td>\n","      <td>5.517230</td>\n","      <td>5.310642</td>\n","      <td>4.927082</td>\n","      <td>7.025551</td>\n","      <td>4.992692</td>\n","      <td>8.006711</td>\n","      <td>7.850989</td>\n","      <td>6.596223</td>\n","      <td>6.962768</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>0</td>\n","      <td>54</td>\n","      <td>5.784453</td>\n","      <td>5.785170</td>\n","      <td>3.978284</td>\n","      <td>4.601635</td>\n","      <td>8.902882</td>\n","      <td>4.107508</td>\n","      <td>4.809921</td>\n","      <td>5.912525</td>\n","      <td>...</td>\n","      <td>5.696565</td>\n","      <td>5.896517</td>\n","      <td>5.604909</td>\n","      <td>5.107001</td>\n","      <td>6.872938</td>\n","      <td>4.533009</td>\n","      <td>8.112374</td>\n","      <td>7.337760</td>\n","      <td>6.405903</td>\n","      <td>7.233815</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0</td>\n","      <td>50</td>\n","      <td>5.283689</td>\n","      <td>5.385669</td>\n","      <td>4.178392</td>\n","      <td>4.617995</td>\n","      <td>8.780384</td>\n","      <td>3.917632</td>\n","      <td>5.144286</td>\n","      <td>6.223132</td>\n","      <td>...</td>\n","      <td>5.053215</td>\n","      <td>5.802501</td>\n","      <td>5.288305</td>\n","      <td>4.732636</td>\n","      <td>7.052588</td>\n","      <td>4.810001</td>\n","      <td>8.001139</td>\n","      <td>7.901643</td>\n","      <td>6.893890</td>\n","      <td>6.624012</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0</td>\n","      <td>31</td>\n","      <td>5.414403</td>\n","      <td>5.828662</td>\n","      <td>3.836844</td>\n","      <td>4.309026</td>\n","      <td>8.221119</td>\n","      <td>4.063579</td>\n","      <td>4.561822</td>\n","      <td>6.135777</td>\n","      <td>...</td>\n","      <td>5.269230</td>\n","      <td>6.576080</td>\n","      <td>5.135130</td>\n","      <td>4.916972</td>\n","      <td>6.706851</td>\n","      <td>4.590855</td>\n","      <td>7.819838</td>\n","      <td>8.041268</td>\n","      <td>6.623575</td>\n","      <td>7.204644</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 17872 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0e9ccda8-0771-4f98-b251-faa0ee942ba0')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-0e9ccda8-0771-4f98-b251-faa0ee942ba0 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-0e9ccda8-0771-4f98-b251-faa0ee942ba0');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-d99b8dd1-792a-4a00-b0bc-3b3314150038\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d99b8dd1-792a-4a00-b0bc-3b3314150038')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-d99b8dd1-792a-4a00-b0bc-3b3314150038 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":8}],"source":["merged_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.234945878Z","start_time":"2023-12-31T21:19:52.231604618Z"},"id":"vTvOiZMsZWQe","outputId":"5a8bf063-04ae-4e58-9082-9a1f9f6295c8","colab":{"base_uri":"https://localhost:8080/","height":273},"executionInfo":{"status":"ok","timestamp":1704801921321,"user_tz":-480,"elapsed":5,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":9,"outputs":[],"source":["# If both age and gender features are available, select 'age' as the condition.\n","condition = 'Age'\n","# Remove the other condition to prevent interference.\n","merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.312540150Z","start_time":"2023-12-31T21:19:52.235772334Z"},"id":"M-4hO0PHZWQe","executionInfo":{"status":"ok","timestamp":1704801926980,"user_tz":-480,"elapsed":738,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":10,"outputs":[],"source":["X = merged_data.drop(columns=[TRAIT, condition]).values\n","Y = merged_data[TRAIT].values\n","Z = merged_data[condition].values"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.323468266Z","start_time":"2023-12-31T21:19:52.290903989Z"},"id":"dhUrhPcqZWQe","executionInfo":{"status":"ok","timestamp":1704801929481,"user_tz":-480,"elapsed":780,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Select the appropriate regression model depending on whether the dataset shows batch effect."],"metadata":{"collapsed":false,"id":"u5Mx3XpAZWQf"}},{"cell_type":"code","execution_count":11,"outputs":[{"output_type":"execute_result","data":{"text/plain":["True"]},"metadata":{},"execution_count":11}],"source":["has_batch_effect = detect_batch_effect(X)\n","has_batch_effect"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.930840782Z","start_time":"2023-12-31T21:19:52.319092572Z"},"id":"GdxnyJr4ZWQg","outputId":"5e6f98fc-40fd-41fc-8f2c-b3d5c86a3926","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704801931525,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":12,"outputs":[],"source":["# Select appropriate models based on whether the dataset has batch effect.\n","# We experiment on two models for each branch. We will decide which one to choose later.\n","\n","if has_batch_effect:\n","    model_constructor1 = VariableSelection\n","    model_params1 = {'modified': True, 'lamda': 3e-4}\n","    model_constructor2 = VariableSelection\n","    model_params2 = {'modified': False}\n","else:\n","    model_constructor1 = Lasso\n","    model_params1 = {'alpha': 1.0, 'random_state': 42}\n","    model_constructor2 = VariableSelection\n","    model_params2 = {'modified': False}"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.931029722Z","start_time":"2023-12-31T21:20:00.929710828Z"},"id":"uzy8ltp-ZWQg","executionInfo":{"status":"ok","timestamp":1704801934573,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["alpha for Lasso: 0.0003\n","alpha for Lasso: 0.0003\n","alpha for Lasso: 0.0003\n","alpha for Lasso: 0.0003\n","alpha for Lasso: 0.0003\n","The cross-validation accuracy is 50.00% ± 17.48%\n"]}],"source":["trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n","cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:20.709459824Z","start_time":"2023-12-31T21:20:00.931657949Z"},"id":"z9tMro7zZWQg","outputId":"0a657205-a43c-43ea-fde2-5224683ec987","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704801968948,"user_tz":-480,"elapsed":529,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","source":["cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"K6gQz2rTGfqt","executionInfo":{"status":"ok","timestamp":1704801998452,"user_tz":-480,"elapsed":24683,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"88da70d8-45e6-4016-8f49-818f1641a88e"},"execution_count":15,"outputs":[{"output_type":"stream","name":"stdout","text":["The cross-validation accuracy is 86.67% ± 6.67%\n"]}]},{"cell_type":"code","execution_count":16,"outputs":[{"output_type":"stream","name":"stdout","text":["alpha for Lasso: 0.0003\n"]}],"source":["normalized_X, _ = normalize_data(X)\n","normalized_Z, _ = normalize_data(Z)\n","\n","# Train regression model on the whole dataset to identify significant genes\n","model1 = ResidualizationRegressor(model_constructor1, model_params1)\n","model1.fit(normalized_X, Y, normalized_Z)\n","\n","model2 = ResidualizationRegressor(model_constructor2, model_params2)\n","model2.fit(normalized_X, Y, normalized_Z)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:24.980870159Z","start_time":"2023-12-31T21:20:20.516710807Z"},"id":"-GzlAnQWZWQg","executionInfo":{"status":"ok","timestamp":1704802010458,"user_tz":-480,"elapsed":5213,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"8b0d2b51-0cc4-4f64-bfd1-62ef2c6c2e50"}},{"cell_type":"markdown","source":["### 5. Discussion and report"],"metadata":{"id":"EjJrxbvb4nlj","ExecuteTime":{"end_time":"2023-10-14T04:50:59.072075075Z","start_time":"2023-10-14T04:50:38.739499998Z"}}},{"cell_type":"code","execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["Effect of the condition on the target variable:\n","Variable: Age\n","Coefficient: 0.3684\n","Found 39 genes with non-zero coefficients associated with the trait 'Sjögren's Syndrome' conditional on the factor 'Age'. These genes are identified as significant based on the regression model.\n","    Variable  Coefficient  Absolute Coefficient\n","     DEFB127    -8.372770              8.372770\n","        RHOC     6.009638              6.009638\n","      RHBDF2     5.878851              5.878851\n","      RAB27B    -4.790219              4.790219\n","    CDC42BPA    -4.549917              4.549917\n","       ITIH1     4.536480              4.536480\n","        SELE     3.277567              3.277567\n","      ACTRT3     2.884623              2.884623\n","       P2RX3     2.527925              2.527925\n","       TMUB1     2.389405              2.389405\n","     ERVK3-2     2.312583              2.312583\n","       GSK3B    -2.208369              2.208369\n","        GFAP     2.068015              2.068015\n","        TPBG    -1.852632              1.852632\n","        C1QC     1.736110              1.736110\n","   SECISBP2L    -1.698493              1.698493\n","       UGGT2     1.539943              1.539943\n","        CUX2     1.511103              1.511103\n","       ADAM7    -1.482144              1.482144\n","     ANKRD36    -1.347046              1.347046\n","     GRIPAP1     1.160157              1.160157\n","     SLC39A8    -0.955279              0.955279\n","     PPP1R3C     0.883244              0.883244\n","     GOLGA7B    -0.847939              0.847939\n","      GARNL3     0.842119              0.842119\n","      ZNF717    -0.811369              0.811369\n","   PRKCQ-AS1    -0.783093              0.783093\n","       HLA-A     0.616969              0.616969\n","LOC101928961     0.605874              0.605874\n","     SLC16A9    -0.487872              0.487872\n","        UBN2    -0.483156              0.483156\n","      MAGEL2    -0.462978              0.462978\n","          C2     0.336528              0.336528\n","        ANK2    -0.199004              0.199004\n","      NKX2-5     0.169477              0.169477\n","         FRK    -0.154183              0.154183\n","   LINC00299     0.134348              0.134348\n","     SLC52A1     0.097907              0.097907\n","       MDGA2     0.034744              0.034744\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/Colab Notebooks/utils.py:434: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n"]}],"source":["feature_cols = merged_data.columns.tolist()\n","feature_cols.remove(TRAIT)\n","\n","threshold = 0.05\n","interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n","                 output_dir=OUTPUT_DIR, model_id=1)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:25.000830410Z","start_time":"2023-12-31T21:20:24.982834541Z"},"id":"FHjsmuPhZWQg","outputId":"d1399ec8-48c1-4451-deb6-07e0bb58bc2c","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704802017509,"user_tz":-480,"elapsed":6,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":18,"outputs":[{"output_type":"stream","name":"stdout","text":["Effect of the condition on the target variable:\n","Variable: Age\n","Coefficient: 0.3684\n","Found 0 significant genes associated with the trait 'Sjögren's Syndrome' conditional on the factor 'Age', with corrected p-value < 0.05:\n","Empty DataFrame\n","Columns: [Variable, Coefficient, p_value, corrected_p_value]\n","Index: []\n"]},{"output_type":"stream","name":"stderr","text":["/content/drive/MyDrive/Colab Notebooks/utils.py:443: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"]}],"source":["interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n","                 output_dir=OUTPUT_DIR, model_id=2)"],"metadata":{"id":"Ia26hwCEZWQg","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704802026714,"user_tz":-480,"elapsed":511,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"710a693f-5558-4477-b93d-d9b0048d623c"}}]}