{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "# Set your preferred name\n",
    "USER = \"Jiayi\"\n",
    "# Set the data and output directories   \n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output'\n",
    "TRAIT = 'Adrenocortical Cancer'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.928143012Z",
     "start_time": "2024-01-10T21:37:25.918685567Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"import os\\nfrom google.colab import drive\\n\\ndrive.mount('/content/drive', force_remount=True)\\nproj_dir = '/content/drive/MyDrive/AI4Science_Public'\\nos.chdir(proj_dir)\""
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.554597756Z",
     "start_time": "2024-01-10T21:37:26.530982646Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.999521704Z",
     "start_time": "2024-01-10T21:37:26.987471072Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Adrenocortical_Cancer_(ACC)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_ACC_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.180581225Z",
     "start_time": "2024-01-10T21:37:27.217363685Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.185775284Z",
     "start_time": "2024-01-10T21:37:30.181745846Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'additional_pharmaceutical_therapy',\n",
       " 'additional_radiation_therapy',\n",
       " 'age_at_initial_pathologic_diagnosis',\n",
       " 'atypical_mitotic_figures',\n",
       " 'bcr_followup_barcode']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.269854441Z",
     "start_time": "2024-01-10T21:37:30.184225525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'age_at_initial_pathologic_diagnosis', 'atypical_mitotic_figures', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'clinical_M', 'ct_scan_findings', 'cytoplasm_presence_less_than_equal_25_percent', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'diffuse_architecture', 'distant_metastasis_anatomic_site', 'excess_adrenal_hormone_diagnosis_method_type', 'excess_adrenal_hormone_history_type', 'form_completion_date', 'gender', 'germline_testing_performed', 'histologic_disease_progression_present_indicator', 'histological_type', 'history_of_neoadjuvant_treatment', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'initial_weight', 'invasion_of_tumor_capsule', 'is_ffpe', 'laterality', 'lost_follow_up', 'lymph_node_examined_count', 'metastatic_neoplasm_confirmed_diagnosis_method_name', 'metastatic_neoplasm_confirmed_diagnosis_method_text', 'mitoses_count', 'mitotane_therapy', 'mitotane_therapy_adjuvant_setting', 'mitotane_therapy_for_macroscopic_residual_disease', 'mitotic_rate', 'necrosis', 'new_neoplasm_confirmed_diagnosis_method_name', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'nuclear_grade_III_IV', 'number_of_lymphnodes_positive_by_he', 'oct_embedded', 'other_dx', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'person_neoplasm_cancer_status', 'post_surgical_procedure_assessment_thyroid_gland_carcinoma_stats', 'postoperative_rx_tx', 'primary_lymph_node_presentation_assessment', 'primary_therapy_outcome_success', 'radiation_therapy', 'residual_tumor', 'ret', 'sample_type', 'sample_type_id', 'sinusoid_invasion', 'therapeutic_mitotane_levels_achieved', 'therapeutic_mitotane_lvl_macroscopic_residual', 'therapeutic_mitotane_lvl_progression', 'therapeutic_mitotane_lvl_recurrence', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'vial_number', 'vital_status', 'weiss_score', 'weiss_venous_invasion', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_ACC_mutation_curated_bcm_gene', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_percentile', '_GENOMIC_ID_data/public/TCGA/ACC/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_ACC_miRNA_HiSeq', '_GENOMIC_ID_TCGA_ACC_RPPA', '_GENOMIC_ID_TCGA_ACC_hMethyl450', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2', '_GENOMIC_ID_TCGA_ACC_gistic2thd', '_GENOMIC_ID_TCGA_ACC_PDMRNAseq', '_GENOMIC_ID_TCGA_ACC_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_ACC_gistic2', '_GENOMIC_ID_TCGA_ACC_mutation_bcm_gene', '_GENOMIC_ID_TCGA_ACC_mutation_curated_broad_gene']\\n\""
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270036796Z",
     "start_time": "2024-01-10T21:37:30.227180075Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = [ 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = [ 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270232386Z",
     "start_time": "2024-01-10T21:37:30.227282564Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'age_at_initial_pathologic_diagnosis': [58, 44, 23, 23, 30],\n",
       " 'days_to_birth': [-21496, -16090, -8624, -8451, -11171],\n",
       " 'year_of_initial_pathologic_diagnosis': [2000, 2004, 2008, 2000, 2000]}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270339062Z",
     "start_time": "2024-01-10T21:37:30.227412883Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270520195Z",
     "start_time": "2024-01-10T21:37:30.227467704Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gender': ['MALE', 'FEMALE', 'FEMALE', 'FEMALE', 'MALE']}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270626721Z",
     "start_time": "2024-01-10T21:37:30.227530279Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270719762Z",
     "start_time": "2024-01-10T21:37:30.227585330Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.001977877Z",
     "start_time": "2024-01-10T21:37:30.227635342Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12 input query terms found dup hits:\t[('GTF2IP1', 2), ('RBMY1A3P', 3), ('RPL31P11', 2), ('HERC2P2', 3), ('WASH3P', 3), ('NUDT9P1', 2), ('\n",
      "154 input query terms found no hit:\t['C16orf13', 'C16orf11', 'LOC100272146', 'LOC339240', 'NACAP1', 'LOC441204', 'KLRA1', 'FAM183A', 'FA\n",
      "10 input query terms found dup hits:\t[('SUGT1P1', 2), ('PTPRVP', 2), ('SNORA62', 3), ('IFITM4P', 7), ('HLA-DRB6', 2), ('FUNDC2P2', 2), ('\n",
      "190 input query terms found no hit:\t['NARFL', 'NFKBIL2', 'LOC150197', 'TMEM84', 'LOC162632', 'PPPDE1', 'PPPDE2', 'C1orf38', 'C1orf31', '\n",
      "11 input query terms found dup hits:\t[('PIP5K1P1', 2), ('HBD', 2), ('PPP1R2P1', 9), ('HSD17B7P2', 2), ('RPSAP9', 2), ('SNORD68', 2), ('SN\n",
      "149 input query terms found no hit:\t['FAM153C', 'C9orf167', 'CLK2P', 'CCDC76', 'CCDC75', 'CCDC72', 'HIST3H2BB', 'PRAC', 'LOC285780', 'LO\n",
      "15 input query terms found dup hits:\t[('SNORD58C', 2), ('UOX', 2), ('UBE2Q2P1', 3), ('PPP4R1L', 2), ('SNORD63', 3), ('ESPNP', 2), ('HBBP1\n",
      "158 input query terms found no hit:\t['C16orf91', 'LOC100130581', 'SFRS2', 'C16orf5', 'C16orf7', 'C16orf3', 'C21orf7', 'C21orf2', 'LOC389\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "15 input query terms found dup hits:\t[('FAM66D', 3), ('FAM66A', 2), ('THSD1P1', 2), ('EEF1DP3', 2), ('PGM5P2', 2), ('UBE2MP1', 2), ('HAR1\n",
      "169 input query terms found no hit:\t['LOC284551', 'LOC285548', 'LOC728410', 'LOC541473', 'DULLARD', 'KIAA0368', 'EFTUD1', 'TWISTNB', 'SF\n",
      "13 input query terms found dup hits:\t[('S100A7L2', 2), ('POM121L8P', 2), ('MEG8', 2), ('KIR3DX1', 5), ('RFPL1S', 2), ('SNORD91B', 2), ('C\n",
      "165 input query terms found no hit:\t['TMEM188', 'PDZD3', 'FAM102B', 'FAM102A', 'SMCR7L', 'G6PC', 'OSTCL', 'LOC653544', 'LOC653545', 'USP\n",
      "16 input query terms found dup hits:\t[('PCNAP1', 2), ('SNORA63', 6), ('SERHL', 2), ('CEACAM22P', 2), ('SNORA16A', 2), ('FAM41AY1', 2), ('\n",
      "147 input query terms found no hit:\t['LRRC37A4', 'LOC100131726', 'CPSF3L', 'COL4A3BP', 'PAR1', 'LOC92973', 'MICALCL', 'SMCR7', 'HIST4H4'\n",
      "15 input query terms found dup hits:\t[('MBL1P', 2), ('SDHAP3', 2), ('PSORS1C3', 8), ('MYADML', 2), ('POM121L10P', 2), ('HLA-J', 9), ('HLA\n",
      "153 input query terms found no hit:\t['LOC100131193', 'KIAA1109', 'DUS2L', 'MYST1', 'IGLL3', 'C12orf69', 'NEURL', 'FAM188B', 'FLJ32063', \n",
      "9 input query terms found dup hits:\t[('PLGLA', 2), ('PA2G4P4', 2), ('RPL23AP7', 2), ('RPL23AP64', 3), ('RPSAP52', 2), ('TPRXL', 2), ('DP\n",
      "177 input query terms found no hit:\t['LOC283314', 'RBM9', 'C20orf118', 'LOC150786', 'C20orf112', 'C20orf111', 'C20orf117', 'C20orf114', \n",
      "18 input query terms found dup hits:\t[('TPTE2P1', 2), ('FOLH1B', 2), ('ID2B', 2), ('MSX2P1', 2), ('RAET1K', 2), ('BMS1P4', 2), ('RPL23AP8\n",
      "175 input query terms found no hit:\t['MYEOV2', 'FAM35B2', 'C20orf85', 'FAM175A', 'C17orf73', 'C17orf72', 'C17orf71', 'C17orf70', 'C17orf\n",
      "19 input query terms found dup hits:\t[('CATSPER2P1', 2), ('AQP7P1', 2), ('ATP8B5P', 2), ('OR7E91P', 2), ('ZNF204P', 2), ('CECR7', 2), ('M\n",
      "144 input query terms found no hit:\t['BASE', 'C14orf21', 'C14orf23', 'FAM5B', 'TROVE2', 'C18orf62', 'LOC401052', 'GNB2L1', 'FAM19A4', 'F\n",
      "13 input query terms found dup hits:\t[('P2RX6P', 2), ('RP9P', 2), ('NRADDP', 2), ('MORF4', 2), ('SNORA75', 6), ('SNORA72', 2), ('FMO6P', \n",
      "168 input query terms found no hit:\t['TTC15', 'C4orf29', 'FAM48A', 'C4orf21', 'C3orf65', 'C4orf23', 'C3orf67', 'LOC144742', 'C3orf63', '\n",
      "19 input query terms found dup hits:\t[('RPL13AP17', 2), ('DNAJB3', 2), ('SNORA58', 2), ('PMCHL1', 2), ('FAM95B1', 2), ('TMEM191A', 2), ('\n",
      "144 input query terms found no hit:\t['LOC254312', 'ZNF542', 'CSDAP1', 'FAM73A', 'FAM73B', 'TCTE3', 'ZNF37B', 'LOC728875', 'KIAA1826', 'P\n",
      "15 input query terms found dup hits:\t[('ZFP91-CNTF', 2), ('PRSS30P', 3), ('GSTM2P1', 2), ('C3P1', 2), ('NCF1B', 2), ('SNORD114-4', 2), ('\n",
      "138 input query terms found no hit:\t['LOC100129716', 'C14orf184', 'C14orf181', 'C14orf183', 'C14orf182', 'C7orf26', 'C7orf27', 'C7orf23'\n",
      "14 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P3', 2), ('ANXA2P2', 2), ('ZNF286B', 2), ('SAA3P', 2), ('GEMIN8P4', 2), ('DI\n",
      "132 input query terms found no hit:\t['AKR1CL1', 'CCDC21', 'CCDC23', 'TMEM111', 'APOB48R', 'LOC595101', 'KIAA1430', 'KIAA1432', 'ORC4L', \n",
      "11 input query terms found dup hits:\t[('TCP10L2', 2), ('FTLP10', 2), ('TPTE2P3', 2), ('VENTXP1', 2), ('XKRY', 2), ('HTR7P1', 2), ('STAG3L\n",
      "175 input query terms found no hit:\t['DFNB59', 'LOC100170939', 'FAM36A', 'C2orf70', 'C2orf71', 'C2orf77', 'C2orf79', 'LOC100128191', 'HR\n",
      "16 input query terms found dup hits:\t[('ZNF702P', 2), ('BRD7P3', 2), ('SCARNA17', 2), ('GABARAPL3', 2), ('MT1IP', 2), ('ADAM6', 3), ('PDZ\n",
      "142 input query terms found no hit:\t['C9orf30', 'PEG3AS', 'C9orf37', 'TXNDC3', 'LOC442308', 'FAM46D', 'FAM46A', 'FAM46B', 'FAM46C', 'WDR\n",
      "10 input query terms found dup hits:\t[('FAM197Y2', 3), ('SNORA71B', 2), ('PRKY', 2), ('NAT8B', 3), ('VENTXP7', 2), ('SPRR2C', 2), ('RPL23\n",
      "148 input query terms found no hit:\t['FCGR1B', 'RG9MTD2', 'RARS', 'LOC727677', 'CCRN4L', 'SEPT14', 'HKR1', 'C1orf112', 'LOC100270710', '\n",
      "16 input query terms found dup hits:\t[('OR6W1P', 3), ('LYPLA2P1', 6), ('RPL23AP53', 2), ('TRPC2', 2), ('RBMY2FP', 2), ('BAGE2', 2), ('POM\n",
      "150 input query terms found no hit:\t['SIP1', 'FAM129C', 'GIF', 'TMCO7', 'CXorf61', 'CXorf64', 'ZNF192', 'SEPP1', 'KIRREL', 'MGC16121', '\n",
      "17 input query terms found dup hits:\t[('ISCA1P1', 2), ('PCDHB19P', 3), ('NAPSB', 2), ('WASH7P', 2), ('TTTY11', 2), ('HCG4', 9), ('OR5E1P'\n",
      "135 input query terms found no hit:\t['PPYR1', 'FAM82A2', 'FLJ37453', 'FAM82A1', 'KIAA0947', 'MOBKL2B', 'MOBKL2C', 'MOBKL2A', 'CMAH', 'GR\n",
      "5 input query terms found dup hits:\t[('GGNBP1', 2), ('WASH5P', 2), ('RPL29P2', 2), ('CDR1', 2), ('OR10J3', 3)]\n",
      "78 input query terms found no hit:\t['C14orf174', 'C14orf176', 'C14orf177', 'C14orf179', 'PTPLB', 'SGEF', 'LOC400752', 'GRLF1', 'SPANXE'\n"
     ]
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-OR-A5LC-01</th>\n",
       "      <th>TCGA-OR-A5JJ-01</th>\n",
       "      <th>TCGA-OR-A5K3-01</th>\n",
       "      <th>TCGA-PK-A5HA-01</th>\n",
       "      <th>TCGA-OR-A5LN-01</th>\n",
       "      <th>TCGA-OR-A5JA-01</th>\n",
       "      <th>TCGA-OR-A5K0-01</th>\n",
       "      <th>TCGA-OR-A5JY-01</th>\n",
       "      <th>TCGA-OR-A5J9-01</th>\n",
       "      <th>TCGA-OR-A5K4-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-OR-A5JT-01</th>\n",
       "      <th>TCGA-OR-A5KW-01</th>\n",
       "      <th>TCGA-OR-A5J8-01</th>\n",
       "      <th>TCGA-OR-A5JQ-01</th>\n",
       "      <th>TCGA-OR-A5JV-01</th>\n",
       "      <th>TCGA-OR-A5KX-01</th>\n",
       "      <th>TCGA-OR-A5L5-01</th>\n",
       "      <th>TCGA-P6-A5OG-01</th>\n",
       "      <th>TCGA-OR-A5LR-01</th>\n",
       "      <th>TCGA-OR-A5LT-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <td>-3.610292</td>\n",
       "      <td>-1.217192</td>\n",
       "      <td>-1.786692</td>\n",
       "      <td>-1.329092</td>\n",
       "      <td>-0.944392</td>\n",
       "      <td>-2.431192</td>\n",
       "      <td>-0.927692</td>\n",
       "      <td>-2.263992</td>\n",
       "      <td>-0.773992</td>\n",
       "      <td>-1.902792</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.863492</td>\n",
       "      <td>-0.979992</td>\n",
       "      <td>-0.217092</td>\n",
       "      <td>-0.166492</td>\n",
       "      <td>-1.806992</td>\n",
       "      <td>-2.295592</td>\n",
       "      <td>-0.664092</td>\n",
       "      <td>-1.707392</td>\n",
       "      <td>-1.648492</td>\n",
       "      <td>0.110608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIF3A</th>\n",
       "      <td>-0.811626</td>\n",
       "      <td>-1.097126</td>\n",
       "      <td>-0.336626</td>\n",
       "      <td>-3.119026</td>\n",
       "      <td>0.294474</td>\n",
       "      <td>-0.000626</td>\n",
       "      <td>-2.902726</td>\n",
       "      <td>-0.038026</td>\n",
       "      <td>-1.558426</td>\n",
       "      <td>6.704374</td>\n",
       "      <td>...</td>\n",
       "      <td>6.097974</td>\n",
       "      <td>-0.775426</td>\n",
       "      <td>1.577074</td>\n",
       "      <td>2.414674</td>\n",
       "      <td>0.083474</td>\n",
       "      <td>-1.224126</td>\n",
       "      <td>4.864274</td>\n",
       "      <td>-0.846526</td>\n",
       "      <td>5.636474</td>\n",
       "      <td>-3.161926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.110765</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.047535</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.346165</td>\n",
       "      <td>0.356765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF10</th>\n",
       "      <td>0.562928</td>\n",
       "      <td>0.398728</td>\n",
       "      <td>1.649328</td>\n",
       "      <td>0.525628</td>\n",
       "      <td>0.649828</td>\n",
       "      <td>0.737528</td>\n",
       "      <td>1.218828</td>\n",
       "      <td>0.940428</td>\n",
       "      <td>1.004428</td>\n",
       "      <td>1.155528</td>\n",
       "      <td>...</td>\n",
       "      <td>1.065328</td>\n",
       "      <td>0.998528</td>\n",
       "      <td>0.649828</td>\n",
       "      <td>0.458028</td>\n",
       "      <td>0.587428</td>\n",
       "      <td>1.318828</td>\n",
       "      <td>1.042828</td>\n",
       "      <td>-0.344672</td>\n",
       "      <td>1.626528</td>\n",
       "      <td>0.328428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF11</th>\n",
       "      <td>-0.735278</td>\n",
       "      <td>-0.698278</td>\n",
       "      <td>-0.527978</td>\n",
       "      <td>-0.639778</td>\n",
       "      <td>-0.482678</td>\n",
       "      <td>-0.101578</td>\n",
       "      <td>-0.736178</td>\n",
       "      <td>0.551822</td>\n",
       "      <td>0.407322</td>\n",
       "      <td>-0.363978</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.211078</td>\n",
       "      <td>-0.070978</td>\n",
       "      <td>-0.112878</td>\n",
       "      <td>-0.053578</td>\n",
       "      <td>0.081922</td>\n",
       "      <td>-0.197478</td>\n",
       "      <td>-0.107178</td>\n",
       "      <td>-0.931178</td>\n",
       "      <td>-0.188578</td>\n",
       "      <td>0.143722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNGT1</th>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-0.749090</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>0.779110</td>\n",
       "      <td>2.326910</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-0.717790</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-0.380790</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TULP3</th>\n",
       "      <td>0.180823</td>\n",
       "      <td>0.425623</td>\n",
       "      <td>-0.333277</td>\n",
       "      <td>-0.350477</td>\n",
       "      <td>0.257623</td>\n",
       "      <td>0.743623</td>\n",
       "      <td>0.424323</td>\n",
       "      <td>0.311723</td>\n",
       "      <td>0.120023</td>\n",
       "      <td>0.279723</td>\n",
       "      <td>...</td>\n",
       "      <td>0.116023</td>\n",
       "      <td>0.111123</td>\n",
       "      <td>1.111123</td>\n",
       "      <td>0.369123</td>\n",
       "      <td>0.040023</td>\n",
       "      <td>-0.731777</td>\n",
       "      <td>-0.038177</td>\n",
       "      <td>-0.328377</td>\n",
       "      <td>0.329723</td>\n",
       "      <td>0.118423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL6B</th>\n",
       "      <td>0.684673</td>\n",
       "      <td>0.075973</td>\n",
       "      <td>-0.968527</td>\n",
       "      <td>0.968073</td>\n",
       "      <td>1.061373</td>\n",
       "      <td>-0.677927</td>\n",
       "      <td>-2.529227</td>\n",
       "      <td>-0.588927</td>\n",
       "      <td>1.577873</td>\n",
       "      <td>-0.527027</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030227</td>\n",
       "      <td>-0.508927</td>\n",
       "      <td>0.993173</td>\n",
       "      <td>0.244273</td>\n",
       "      <td>0.462973</td>\n",
       "      <td>0.735173</td>\n",
       "      <td>0.129673</td>\n",
       "      <td>3.400573</td>\n",
       "      <td>0.408673</td>\n",
       "      <td>-2.482227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSTK1</th>\n",
       "      <td>1.219405</td>\n",
       "      <td>1.561405</td>\n",
       "      <td>2.331205</td>\n",
       "      <td>1.034305</td>\n",
       "      <td>1.308005</td>\n",
       "      <td>0.183105</td>\n",
       "      <td>0.528505</td>\n",
       "      <td>1.042305</td>\n",
       "      <td>1.334905</td>\n",
       "      <td>-3.908095</td>\n",
       "      <td>...</td>\n",
       "      <td>0.780705</td>\n",
       "      <td>0.853105</td>\n",
       "      <td>0.093105</td>\n",
       "      <td>1.335505</td>\n",
       "      <td>1.161405</td>\n",
       "      <td>-1.075095</td>\n",
       "      <td>0.658005</td>\n",
       "      <td>0.240205</td>\n",
       "      <td>0.236505</td>\n",
       "      <td>-0.423795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELP</th>\n",
       "      <td>-3.390633</td>\n",
       "      <td>-3.333433</td>\n",
       "      <td>-2.624733</td>\n",
       "      <td>-0.030433</td>\n",
       "      <td>-3.797833</td>\n",
       "      <td>-5.525533</td>\n",
       "      <td>-4.913433</td>\n",
       "      <td>-4.991633</td>\n",
       "      <td>-3.630333</td>\n",
       "      <td>-4.441133</td>\n",
       "      <td>...</td>\n",
       "      <td>1.152367</td>\n",
       "      <td>-1.514633</td>\n",
       "      <td>-2.186733</td>\n",
       "      <td>0.956467</td>\n",
       "      <td>-1.081333</td>\n",
       "      <td>-2.124933</td>\n",
       "      <td>2.108967</td>\n",
       "      <td>0.190967</td>\n",
       "      <td>0.199167</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17339 rows Ã— 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TCGA-OR-A5LC-01  TCGA-OR-A5JJ-01  TCGA-OR-A5K3-01  TCGA-PK-A5HA-01  \\\n",
       "ARHGEF10L        -3.610292        -1.217192        -1.786692        -1.329092   \n",
       "HIF3A            -0.811626        -1.097126        -0.336626        -3.119026   \n",
       "RNF17            -0.531035        -0.531035        -0.531035        -0.531035   \n",
       "RNF10             0.562928         0.398728         1.649328         0.525628   \n",
       "RNF11            -0.735278        -0.698278        -0.527978        -0.639778   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -1.281390        -0.749090        -1.281390        -1.281390   \n",
       "TULP3             0.180823         0.425623        -0.333277        -0.350477   \n",
       "BCL6B             0.684673         0.075973        -0.968527         0.968073   \n",
       "GSTK1             1.219405         1.561405         2.331205         1.034305   \n",
       "SELP             -3.390633        -3.333433        -2.624733        -0.030433   \n",
       "\n",
       "           TCGA-OR-A5LN-01  TCGA-OR-A5JA-01  TCGA-OR-A5K0-01  TCGA-OR-A5JY-01  \\\n",
       "ARHGEF10L        -0.944392        -2.431192        -0.927692        -2.263992   \n",
       "HIF3A             0.294474        -0.000626        -2.902726        -0.038026   \n",
       "RNF17            -0.531035        -0.531035        -0.531035        -0.531035   \n",
       "RNF10             0.649828         0.737528         1.218828         0.940428   \n",
       "RNF11            -0.482678        -0.101578        -0.736178         0.551822   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -1.281390        -1.281390         0.779110         2.326910   \n",
       "TULP3             0.257623         0.743623         0.424323         0.311723   \n",
       "BCL6B             1.061373        -0.677927        -2.529227        -0.588927   \n",
       "GSTK1             1.308005         0.183105         0.528505         1.042305   \n",
       "SELP             -3.797833        -5.525533        -4.913433        -4.991633   \n",
       "\n",
       "           TCGA-OR-A5J9-01  TCGA-OR-A5K4-01  ...  TCGA-OR-A5JT-01  \\\n",
       "ARHGEF10L        -0.773992        -1.902792  ...        -0.863492   \n",
       "HIF3A            -1.558426         6.704374  ...         6.097974   \n",
       "RNF17            -0.531035         0.110765  ...        -0.531035   \n",
       "RNF10             1.004428         1.155528  ...         1.065328   \n",
       "RNF11             0.407322        -0.363978  ...        -0.211078   \n",
       "...                    ...              ...  ...              ...   \n",
       "GNGT1            -1.281390        -1.281390  ...        -1.281390   \n",
       "TULP3             0.120023         0.279723  ...         0.116023   \n",
       "BCL6B             1.577873        -0.527027  ...        -0.030227   \n",
       "GSTK1             1.334905        -3.908095  ...         0.780705   \n",
       "SELP             -3.630333        -4.441133  ...         1.152367   \n",
       "\n",
       "           TCGA-OR-A5KW-01  TCGA-OR-A5J8-01  TCGA-OR-A5JQ-01  TCGA-OR-A5JV-01  \\\n",
       "ARHGEF10L        -0.979992        -0.217092        -0.166492        -1.806992   \n",
       "HIF3A            -0.775426         1.577074         2.414674         0.083474   \n",
       "RNF17            -0.531035        -0.531035        -0.047535        -0.531035   \n",
       "RNF10             0.998528         0.649828         0.458028         0.587428   \n",
       "RNF11            -0.070978        -0.112878        -0.053578         0.081922   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -1.281390        -1.281390        -1.281390        -1.281390   \n",
       "TULP3             0.111123         1.111123         0.369123         0.040023   \n",
       "BCL6B            -0.508927         0.993173         0.244273         0.462973   \n",
       "GSTK1             0.853105         0.093105         1.335505         1.161405   \n",
       "SELP             -1.514633        -2.186733         0.956467        -1.081333   \n",
       "\n",
       "           TCGA-OR-A5KX-01  TCGA-OR-A5L5-01  TCGA-P6-A5OG-01  TCGA-OR-A5LR-01  \\\n",
       "ARHGEF10L        -2.295592        -0.664092        -1.707392        -1.648492   \n",
       "HIF3A            -1.224126         4.864274        -0.846526         5.636474   \n",
       "RNF17            -0.531035        -0.531035        -0.531035         0.346165   \n",
       "RNF10             1.318828         1.042828        -0.344672         1.626528   \n",
       "RNF11            -0.197478        -0.107178        -0.931178        -0.188578   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -0.717790        -1.281390        -0.380790        -1.281390   \n",
       "TULP3            -0.731777        -0.038177        -0.328377         0.329723   \n",
       "BCL6B             0.735173         0.129673         3.400573         0.408673   \n",
       "GSTK1            -1.075095         0.658005         0.240205         0.236505   \n",
       "SELP             -2.124933         2.108967         0.190967         0.199167   \n",
       "\n",
       "           TCGA-OR-A5LT-01  \n",
       "ARHGEF10L         0.110608  \n",
       "HIF3A            -3.161926  \n",
       "RNF17             0.356765  \n",
       "RNF10             0.328428  \n",
       "RNF11             0.143722  \n",
       "...                    ...  \n",
       "GNGT1            -1.281390  \n",
       "TULP3             0.118423  \n",
       "BCL6B            -2.482227  \n",
       "GSTK1            -0.423795  \n",
       "SELP             -5.525533  \n",
       "\n",
       "[17339 rows x 79 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.108889749Z",
     "start_time": "2024-01-10T21:38:18.021199718Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.641092</td>\n",
       "      <td>-0.325826</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.266428</td>\n",
       "      <td>0.355422</td>\n",
       "      <td>0.03719</td>\n",
       "      <td>0.243706</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.520186</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.182978</td>\n",
       "      <td>-0.615817</td>\n",
       "      <td>-0.281533</td>\n",
       "      <td>3.02111</td>\n",
       "      <td>-0.927577</td>\n",
       "      <td>-1.006227</td>\n",
       "      <td>1.119905</td>\n",
       "      <td>-2.185533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.864792</td>\n",
       "      <td>2.766674</td>\n",
       "      <td>0.321165</td>\n",
       "      <td>1.000728</td>\n",
       "      <td>0.836122</td>\n",
       "      <td>0.35439</td>\n",
       "      <td>-0.436694</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.318586</td>\n",
       "      <td>1.056018</td>\n",
       "      <td>0.393822</td>\n",
       "      <td>2.366583</td>\n",
       "      <td>-0.955033</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>1.020723</td>\n",
       "      <td>1.226373</td>\n",
       "      <td>1.164005</td>\n",
       "      <td>0.265067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.723192</td>\n",
       "      <td>-0.362926</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.639828</td>\n",
       "      <td>-0.199578</td>\n",
       "      <td>-0.48331</td>\n",
       "      <td>0.143606</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.574486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>-0.113317</td>\n",
       "      <td>-3.803333</td>\n",
       "      <td>-0.61009</td>\n",
       "      <td>0.397623</td>\n",
       "      <td>-0.675227</td>\n",
       "      <td>1.196005</td>\n",
       "      <td>-3.161633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.576792</td>\n",
       "      <td>-2.086226</td>\n",
       "      <td>2.463765</td>\n",
       "      <td>1.382228</td>\n",
       "      <td>-1.115678</td>\n",
       "      <td>-1.23621</td>\n",
       "      <td>0.615806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.279486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.078622</td>\n",
       "      <td>1.095983</td>\n",
       "      <td>-0.908533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>0.661823</td>\n",
       "      <td>0.458273</td>\n",
       "      <td>0.839605</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J6-01</th>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.311992</td>\n",
       "      <td>5.225974</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.967928</td>\n",
       "      <td>-0.393778</td>\n",
       "      <td>-0.38231</td>\n",
       "      <td>-0.060194</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>1.607218</td>\n",
       "      <td>2.481122</td>\n",
       "      <td>-0.946617</td>\n",
       "      <td>-0.570533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.425177</td>\n",
       "      <td>0.938573</td>\n",
       "      <td>0.495005</td>\n",
       "      <td>-1.733333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Adrenocortical Cancer  Age  Gender  ARHGEF10L     HIF3A  \\\n",
       "sampleID                                                                   \n",
       "TCGA-OR-A5J1-01                      1   58       1  -0.641092 -0.325826   \n",
       "TCGA-OR-A5J2-01                      1   44       0  -1.864792  2.766674   \n",
       "TCGA-OR-A5J3-01                      1   23       0  -0.723192 -0.362926   \n",
       "TCGA-OR-A5J5-01                      1   30       1  -1.576792 -2.086226   \n",
       "TCGA-OR-A5J6-01                      1   29       0  -2.311992  5.225974   \n",
       "\n",
       "                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n",
       "sampleID                                                          ...   \n",
       "TCGA-OR-A5J1-01 -0.531035  1.266428  0.355422  0.03719  0.243706  ...   \n",
       "TCGA-OR-A5J2-01  0.321165  1.000728  0.836122  0.35439 -0.436694  ...   \n",
       "TCGA-OR-A5J3-01 -0.531035  0.639828 -0.199578 -0.48331  0.143606  ...   \n",
       "TCGA-OR-A5J5-01  2.463765  1.382228 -1.115678 -1.23621  0.615806  ...   \n",
       "TCGA-OR-A5J6-01 -0.531035  0.967928 -0.393778 -0.38231 -0.060194  ...   \n",
       "\n",
       "                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n",
       "sampleID                                                                     \n",
       "TCGA-OR-A5J1-01 -1.520186 -0.086682 -0.182978 -0.615817 -0.281533  3.02111   \n",
       "TCGA-OR-A5J2-01 -0.318586  1.056018  0.393822  2.366583 -0.955033 -1.28139   \n",
       "TCGA-OR-A5J3-01 -0.574486 -0.086682 -0.748878 -0.113317 -3.803333 -0.61009   \n",
       "TCGA-OR-A5J5-01 -0.279486 -0.086682  0.078622  1.095983 -0.908533 -1.28139   \n",
       "TCGA-OR-A5J6-01 -2.090786  1.607218  2.481122 -0.946617 -0.570533 -1.28139   \n",
       "\n",
       "                    TULP3     BCL6B     GSTK1      SELP  \n",
       "sampleID                                                 \n",
       "TCGA-OR-A5J1-01 -0.927577 -1.006227  1.119905 -2.185533  \n",
       "TCGA-OR-A5J2-01  1.020723  1.226373  1.164005  0.265067  \n",
       "TCGA-OR-A5J3-01  0.397623 -0.675227  1.196005 -3.161633  \n",
       "TCGA-OR-A5J5-01  0.661823  0.458273  0.839605 -5.525533  \n",
       "TCGA-OR-A5J6-01 -0.425177  0.938573  0.495005 -1.733333  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.641092</td>\n",
       "      <td>-0.325826</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.266428</td>\n",
       "      <td>0.355422</td>\n",
       "      <td>0.03719</td>\n",
       "      <td>0.243706</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.520186</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.182978</td>\n",
       "      <td>-0.615817</td>\n",
       "      <td>-0.281533</td>\n",
       "      <td>3.02111</td>\n",
       "      <td>-0.927577</td>\n",
       "      <td>-1.006227</td>\n",
       "      <td>1.119905</td>\n",
       "      <td>-2.185533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.864792</td>\n",
       "      <td>2.766674</td>\n",
       "      <td>0.321165</td>\n",
       "      <td>1.000728</td>\n",
       "      <td>0.836122</td>\n",
       "      <td>0.35439</td>\n",
       "      <td>-0.436694</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.318586</td>\n",
       "      <td>1.056018</td>\n",
       "      <td>0.393822</td>\n",
       "      <td>2.366583</td>\n",
       "      <td>-0.955033</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>1.020723</td>\n",
       "      <td>1.226373</td>\n",
       "      <td>1.164005</td>\n",
       "      <td>0.265067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.723192</td>\n",
       "      <td>-0.362926</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.639828</td>\n",
       "      <td>-0.199578</td>\n",
       "      <td>-0.48331</td>\n",
       "      <td>0.143606</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.574486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>-0.113317</td>\n",
       "      <td>-3.803333</td>\n",
       "      <td>-0.61009</td>\n",
       "      <td>0.397623</td>\n",
       "      <td>-0.675227</td>\n",
       "      <td>1.196005</td>\n",
       "      <td>-3.161633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.576792</td>\n",
       "      <td>-2.086226</td>\n",
       "      <td>2.463765</td>\n",
       "      <td>1.382228</td>\n",
       "      <td>-1.115678</td>\n",
       "      <td>-1.23621</td>\n",
       "      <td>0.615806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.279486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.078622</td>\n",
       "      <td>1.095983</td>\n",
       "      <td>-0.908533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>0.661823</td>\n",
       "      <td>0.458273</td>\n",
       "      <td>0.839605</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J6-01</th>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.311992</td>\n",
       "      <td>5.225974</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.967928</td>\n",
       "      <td>-0.393778</td>\n",
       "      <td>-0.38231</td>\n",
       "      <td>-0.060194</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>1.607218</td>\n",
       "      <td>2.481122</td>\n",
       "      <td>-0.946617</td>\n",
       "      <td>-0.570533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.425177</td>\n",
       "      <td>0.938573</td>\n",
       "      <td>0.495005</td>\n",
       "      <td>-1.733333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PA-A5YG-01</th>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "      <td>0.189908</td>\n",
       "      <td>1.377574</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.798828</td>\n",
       "      <td>0.174722</td>\n",
       "      <td>0.92979</td>\n",
       "      <td>-0.043394</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.542586</td>\n",
       "      <td>0.857918</td>\n",
       "      <td>0.195722</td>\n",
       "      <td>1.314483</td>\n",
       "      <td>0.393367</td>\n",
       "      <td>-0.73319</td>\n",
       "      <td>-0.104177</td>\n",
       "      <td>0.280873</td>\n",
       "      <td>2.095805</td>\n",
       "      <td>1.767267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PK-A5H8-01</th>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>-3.279892</td>\n",
       "      <td>3.689074</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.198428</td>\n",
       "      <td>-0.118078</td>\n",
       "      <td>-0.39651</td>\n",
       "      <td>0.470906</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.390686</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.048778</td>\n",
       "      <td>-1.587117</td>\n",
       "      <td>-1.890433</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.098077</td>\n",
       "      <td>-0.338827</td>\n",
       "      <td>1.313105</td>\n",
       "      <td>-3.278333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PK-A5H9-01</th>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0.724908</td>\n",
       "      <td>4.032174</td>\n",
       "      <td>-0.099235</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.183822</td>\n",
       "      <td>0.16789</td>\n",
       "      <td>1.447606</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.052486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>1.641683</td>\n",
       "      <td>-1.610433</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.707577</td>\n",
       "      <td>1.106373</td>\n",
       "      <td>0.321005</td>\n",
       "      <td>0.311967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PK-A5HA-01</th>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.329092</td>\n",
       "      <td>-3.119026</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.525628</td>\n",
       "      <td>-0.639778</td>\n",
       "      <td>-0.04121</td>\n",
       "      <td>0.112606</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.115378</td>\n",
       "      <td>-1.587117</td>\n",
       "      <td>-0.018633</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.350477</td>\n",
       "      <td>0.968073</td>\n",
       "      <td>1.034305</td>\n",
       "      <td>-0.030433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-PK-A5HB-01</th>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>1</td>\n",
       "      <td>0.592508</td>\n",
       "      <td>-1.430426</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.901628</td>\n",
       "      <td>0.795722</td>\n",
       "      <td>-0.95351</td>\n",
       "      <td>-0.232294</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.241286</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.262378</td>\n",
       "      <td>-1.100617</td>\n",
       "      <td>-2.149933</td>\n",
       "      <td>-0.43189</td>\n",
       "      <td>-0.360377</td>\n",
       "      <td>-0.661627</td>\n",
       "      <td>-0.329495</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows Ã— 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Adrenocortical Cancer  Age  Gender  ARHGEF10L     HIF3A  \\\n",
       "sampleID                                                                   \n",
       "TCGA-OR-A5J1-01                      1   58       1  -0.641092 -0.325826   \n",
       "TCGA-OR-A5J2-01                      1   44       0  -1.864792  2.766674   \n",
       "TCGA-OR-A5J3-01                      1   23       0  -0.723192 -0.362926   \n",
       "TCGA-OR-A5J5-01                      1   30       1  -1.576792 -2.086226   \n",
       "TCGA-OR-A5J6-01                      1   29       0  -2.311992  5.225974   \n",
       "...                                ...  ...     ...        ...       ...   \n",
       "TCGA-PA-A5YG-01                      1   51       1   0.189908  1.377574   \n",
       "TCGA-PK-A5H8-01                      1   42       1  -3.279892  3.689074   \n",
       "TCGA-PK-A5H9-01                      1   27       0   0.724908  4.032174   \n",
       "TCGA-PK-A5HA-01                      1   63       1  -1.329092 -3.119026   \n",
       "TCGA-PK-A5HB-01                      1   63       1   0.592508 -1.430426   \n",
       "\n",
       "                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n",
       "sampleID                                                          ...   \n",
       "TCGA-OR-A5J1-01 -0.531035  1.266428  0.355422  0.03719  0.243706  ...   \n",
       "TCGA-OR-A5J2-01  0.321165  1.000728  0.836122  0.35439 -0.436694  ...   \n",
       "TCGA-OR-A5J3-01 -0.531035  0.639828 -0.199578 -0.48331  0.143606  ...   \n",
       "TCGA-OR-A5J5-01  2.463765  1.382228 -1.115678 -1.23621  0.615806  ...   \n",
       "TCGA-OR-A5J6-01 -0.531035  0.967928 -0.393778 -0.38231 -0.060194  ...   \n",
       "...                   ...       ...       ...      ...       ...  ...   \n",
       "TCGA-PA-A5YG-01 -0.531035  0.798828  0.174722  0.92979 -0.043394  ...   \n",
       "TCGA-PK-A5H8-01 -0.531035  1.198428 -0.118078 -0.39651  0.470906  ...   \n",
       "TCGA-PK-A5H9-01 -0.099235  0.344828  0.183822  0.16789  1.447606  ...   \n",
       "TCGA-PK-A5HA-01 -0.531035  0.525628 -0.639778 -0.04121  0.112606  ...   \n",
       "TCGA-PK-A5HB-01 -0.531035  0.901628  0.795722 -0.95351 -0.232294  ...   \n",
       "\n",
       "                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n",
       "sampleID                                                                     \n",
       "TCGA-OR-A5J1-01 -1.520186 -0.086682 -0.182978 -0.615817 -0.281533  3.02111   \n",
       "TCGA-OR-A5J2-01 -0.318586  1.056018  0.393822  2.366583 -0.955033 -1.28139   \n",
       "TCGA-OR-A5J3-01 -0.574486 -0.086682 -0.748878 -0.113317 -3.803333 -0.61009   \n",
       "TCGA-OR-A5J5-01 -0.279486 -0.086682  0.078622  1.095983 -0.908533 -1.28139   \n",
       "TCGA-OR-A5J6-01 -2.090786  1.607218  2.481122 -0.946617 -0.570533 -1.28139   \n",
       "...                   ...       ...       ...       ...       ...      ...   \n",
       "TCGA-PA-A5YG-01 -1.542586  0.857918  0.195722  1.314483  0.393367 -0.73319   \n",
       "TCGA-PK-A5H8-01 -1.390686 -0.086682 -0.048778 -1.587117 -1.890433 -1.28139   \n",
       "TCGA-PK-A5H9-01 -1.052486 -0.086682 -0.748878  1.641683 -1.610433 -1.28139   \n",
       "TCGA-PK-A5HA-01 -2.090786 -0.086682 -0.115378 -1.587117 -0.018633 -1.28139   \n",
       "TCGA-PK-A5HB-01 -1.241286 -0.086682 -0.262378 -1.100617 -2.149933 -0.43189   \n",
       "\n",
       "                    TULP3     BCL6B     GSTK1      SELP  \n",
       "sampleID                                                 \n",
       "TCGA-OR-A5J1-01 -0.927577 -1.006227  1.119905 -2.185533  \n",
       "TCGA-OR-A5J2-01  1.020723  1.226373  1.164005  0.265067  \n",
       "TCGA-OR-A5J3-01  0.397623 -0.675227  1.196005 -3.161633  \n",
       "TCGA-OR-A5J5-01  0.661823  0.458273  0.839605 -5.525533  \n",
       "TCGA-OR-A5J6-01 -0.425177  0.938573  0.495005 -1.733333  \n",
       "...                   ...       ...       ...       ...  \n",
       "TCGA-PA-A5YG-01 -0.104177  0.280873  2.095805  1.767267  \n",
       "TCGA-PK-A5H8-01 -0.098077 -0.338827  1.313105 -3.278333  \n",
       "TCGA-PK-A5H9-01 -0.707577  1.106373  0.321005  0.311967  \n",
       "TCGA-PK-A5HA-01 -0.350477  0.968073  1.034305 -0.030433  \n",
       "TCGA-PK-A5HB-01 -0.360377 -0.661627 -0.329495 -5.525533  \n",
       "\n",
       "[79 rows x 17342 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.112793384Z",
     "start_time": "2024-01-10T21:38:18.107415526Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 79 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1' with 79 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 35.0\n",
      "  50% (Median): 49.0\n",
      "  75%: 59.5\n",
      "Min: 14\n",
      "Max: 77\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1' with 31 occurrences. This represents 39.24% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.377898533Z",
     "start_time": "2024-01-10T21:38:18.111417452Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425336629Z",
     "start_time": "2024-01-10T21:38:31.419043732Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE108088',\n",
       " 'GSE108089',\n",
       " 'GSE143383',\n",
       " 'GSE169253',\n",
       " 'GSE19750',\n",
       " 'GSE19776',\n",
       " 'GSE19856',\n",
       " 'GSE21660',\n",
       " 'GSE32206',\n",
       " 'GSE33371',\n",
       " 'GSE35066',\n",
       " 'GSE36353',\n",
       " 'GSE49276',\n",
       " 'GSE49277',\n",
       " 'GSE49278',\n",
       " 'GSE49280',\n",
       " 'GSE52296',\n",
       " 'GSE67766',\n",
       " 'GSE68606',\n",
       " 'GSE68950',\n",
       " 'GSE75415',\n",
       " 'GSE76019',\n",
       " 'GSE90713']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Adrenocortical-Cancer\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:02.935079426Z",
     "start_time": "2024-01-10T21:39:02.923698385Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Beta-catenin status effects in human adrenocortical carcinomas (33), adenomas (22), and normal adrenal cortex (10)\"\n",
      "!Series_summary\t\"We scored adrenocortical carcinomas and adenomas for abnormal beta-catenin staining, and sequenced the beta-catenin gene in some samples. We compared adrenocortincal carcinomas with and without abnormal beta-catenin staining and found many significant expression differences and significant results from enrichment testing. A similar comparison in the adenomas gave relatively few differences, and they did not correlate to differences found for the carcinomas.  Abnormal beta-catenin staining was associated with mitotic rate and poorer patient survival in the carcinomas.  In a second independent data set (given in a supplement) we again found beta-catenin associated with poor survival.  The array data given is the same as GEO series GSE10927, with additional characteristics about beta-catenin, and new patient followup data.  The analysis shown in a supplementary Excel file is also new.\"\n",
      "!Series_overall_design\t\"Human samples of 33 adrenocortical carcinomas, 22 adrenocortical adenomas, and 10 normal adrenal cortex samples, each from a different patient, had mRNA assays performed using Affymetrix HG_U133_plus_2 arrays, with 54675 probe-sets.  23 of the carcinomas were used in survival analysis.  27 additional samples without array data were also used in a survival analysis.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM825367</th>\n",
       "      <th>GSM825368</th>\n",
       "      <th>GSM825369</th>\n",
       "      <th>GSM825370</th>\n",
       "      <th>GSM825371</th>\n",
       "      <th>GSM825372</th>\n",
       "      <th>GSM825373</th>\n",
       "      <th>GSM825374</th>\n",
       "      <th>GSM825375</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM825422</th>\n",
       "      <th>GSM825423</th>\n",
       "      <th>GSM825424</th>\n",
       "      <th>GSM825425</th>\n",
       "      <th>GSM825426</th>\n",
       "      <th>GSM825427</th>\n",
       "      <th>GSM825428</th>\n",
       "      <th>GSM825429</th>\n",
       "      <th>GSM825430</th>\n",
       "      <th>GSM825431</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 58</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 44</td>\n",
       "      <td>age: 32</td>\n",
       "      <td>age: 28</td>\n",
       "      <td>age: 55</td>\n",
       "      <td>age: 78</td>\n",
       "      <td>age: 41</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 19</td>\n",
       "      <td>age: 61</td>\n",
       "      <td>age: 33</td>\n",
       "      <td>age: 31</td>\n",
       "      <td>age: 60</td>\n",
       "      <td>age: 51</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 25</td>\n",
       "      <td>age: 47</td>\n",
       "      <td>age: 49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Left</td>\n",
       "      <td>side of body: not applicable</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Right</td>\n",
       "      <td>side of body: Left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for me...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for AC...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for re...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for AC...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for ACA</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for ph...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for mu...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for re...</td>\n",
       "      <td>clinical characteristics: Adrenalectomy for re...</td>\n",
       "      <td>...</td>\n",
       "      <td>clinical characteristics: Left adrenal mass</td>\n",
       "      <td>clinical characteristics: Cushing Syndrome</td>\n",
       "      <td>clinical characteristics: Cortisol secreting t...</td>\n",
       "      <td>clinical characteristics: History of Adrenocor...</td>\n",
       "      <td>clinical characteristics: Cushing Syndrome</td>\n",
       "      <td>clinical characteristics: Right adrenal mass; ...</td>\n",
       "      <td>clinical characteristics: Right adrenal tumor;...</td>\n",
       "      <td>clinical characteristics: History of adrenal m...</td>\n",
       "      <td>clinical characteristics: Cushing Syndrome; Ri...</td>\n",
       "      <td>clinical characteristics: Elevated serum aldos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>tumor diameter (cm): not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor diameter (cm): 10.5</td>\n",
       "      <td>tumor diameter (cm): 9.8</td>\n",
       "      <td>tumor diameter (cm): 13</td>\n",
       "      <td>tumor diameter (cm): unknown</td>\n",
       "      <td>tumor diameter (cm): 18.5</td>\n",
       "      <td>tumor diameter (cm): 15.5</td>\n",
       "      <td>tumor diameter (cm): 22</td>\n",
       "      <td>tumor diameter (cm): 26</td>\n",
       "      <td>tumor diameter (cm): unknown</td>\n",
       "      <td>tumor diameter (cm): 8.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>tumor weight (gm): not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor weight (gm): unknown</td>\n",
       "      <td>tumor weight (gm): 253</td>\n",
       "      <td>tumor weight (gm): 660</td>\n",
       "      <td>tumor weight (gm): unknown</td>\n",
       "      <td>tumor weight (gm): 1805</td>\n",
       "      <td>tumor weight (gm): 1305</td>\n",
       "      <td>tumor weight (gm): 2440</td>\n",
       "      <td>tumor weight (gm): 2300</td>\n",
       "      <td>tumor weight (gm): unknown</td>\n",
       "      <td>tumor weight (gm): 397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>weiss score of tumor: not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: Low</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "      <td>weiss score of tumor: High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>mitotic rate of tumor: not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>mitotic rate of tumor: 27</td>\n",
       "      <td>mitotic rate of tumor: 27</td>\n",
       "      <td>mitotic rate of tumor: 70</td>\n",
       "      <td>mitotic rate of tumor: 16</td>\n",
       "      <td>mitotic rate of tumor: 40</td>\n",
       "      <td>mitotic rate of tumor: 23</td>\n",
       "      <td>mitotic rate of tumor: 22</td>\n",
       "      <td>mitotic rate of tumor: 34</td>\n",
       "      <td>mitotic rate of tumor: 37</td>\n",
       "      <td>mitotic rate of tumor: 28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor stage: 2</td>\n",
       "      <td>tumor stage: 2</td>\n",
       "      <td>tumor stage: 3</td>\n",
       "      <td>tumor stage: 4</td>\n",
       "      <td>tumor stage: 2</td>\n",
       "      <td>tumor stage: 3</td>\n",
       "      <td>tumor stage: 4</td>\n",
       "      <td>tumor stage: 4</td>\n",
       "      <td>tumor stage: 1</td>\n",
       "      <td>tumor stage: 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>years to last followup: 0.619178082</td>\n",
       "      <td>years to last followup: 0.942465753</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "      <td>years to last followup: 0.364383562</td>\n",
       "      <td>years to last followup: 2.232876712</td>\n",
       "      <td>years to last followup: 0.624657534</td>\n",
       "      <td>years to last followup: 0.898630137</td>\n",
       "      <td>years to last followup: 7.81</td>\n",
       "      <td>years to last followup: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>dead or alive at last followup: dead</td>\n",
       "      <td>dead or alive at last followup: dead</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "      <td>dead or alive at last followup: dead</td>\n",
       "      <td>dead or alive at last followup: dead</td>\n",
       "      <td>dead or alive at last followup: dead</td>\n",
       "      <td>dead or alive at last followup: dead</td>\n",
       "      <td>dead or alive at last followup: alive</td>\n",
       "      <td>dead or alive at last followup: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>beta-catenin staining: Membrane</td>\n",
       "      <td>beta-catenin staining: unknown</td>\n",
       "      <td>beta-catenin staining: unknown</td>\n",
       "      <td>beta-catenin staining: unknown</td>\n",
       "      <td>beta-catenin staining: Membrane</td>\n",
       "      <td>beta-catenin staining: Membrane</td>\n",
       "      <td>beta-catenin staining: Membrane</td>\n",
       "      <td>beta-catenin staining: unknown</td>\n",
       "      <td>beta-catenin staining: unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: Nuclear</td>\n",
       "      <td>beta-catenin staining: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "      <td>beta-catenin sequence result: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>hormone expression: unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>hormone expression: Cortisol</td>\n",
       "      <td>hormone expression: Cortisol</td>\n",
       "      <td>hormone expression: Cortisol</td>\n",
       "      <td>hormone expression: No_information</td>\n",
       "      <td>hormone expression: Cortisol</td>\n",
       "      <td>hormone expression: Testosterone</td>\n",
       "      <td>hormone expression: Non-functional</td>\n",
       "      <td>hormone expression: Cortisol+Testosterone</td>\n",
       "      <td>hormone expression: Cortisol</td>\n",
       "      <td>hormone expression: No_information</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows Ã— 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          !Sample_geo_accession  \\\n",
       "0   !Sample_characteristics_ch1   \n",
       "1   !Sample_characteristics_ch1   \n",
       "2   !Sample_characteristics_ch1   \n",
       "3   !Sample_characteristics_ch1   \n",
       "4   !Sample_characteristics_ch1   \n",
       "5   !Sample_characteristics_ch1   \n",
       "6   !Sample_characteristics_ch1   \n",
       "7   !Sample_characteristics_ch1   \n",
       "8   !Sample_characteristics_ch1   \n",
       "9   !Sample_characteristics_ch1   \n",
       "10  !Sample_characteristics_ch1   \n",
       "11  !Sample_characteristics_ch1   \n",
       "12  !Sample_characteristics_ch1   \n",
       "13  !Sample_characteristics_ch1   \n",
       "\n",
       "                                            GSM825367  \\\n",
       "0                                             age: 71   \n",
       "1                                              Sex: F   \n",
       "2                                  side of body: Left   \n",
       "3   clinical characteristics: Adrenalectomy for me...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                    beta-catenin staining: Membrane   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                            GSM825368  \\\n",
       "0                                             age: 58   \n",
       "1                                              Sex: F   \n",
       "2                                 side of body: Right   \n",
       "3   clinical characteristics: Adrenalectomy for AC...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                     beta-catenin staining: unknown   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                            GSM825369  \\\n",
       "0                                             age: 71   \n",
       "1                                              Sex: M   \n",
       "2                                 side of body: Right   \n",
       "3   clinical characteristics: Adrenalectomy for re...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                     beta-catenin staining: unknown   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                            GSM825370  \\\n",
       "0                                             age: 44   \n",
       "1                                              Sex: M   \n",
       "2                                  side of body: Left   \n",
       "3   clinical characteristics: Adrenalectomy for AC...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                     beta-catenin staining: unknown   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                          GSM825371  \\\n",
       "0                                           age: 32   \n",
       "1                                            Sex: F   \n",
       "2                               side of body: Right   \n",
       "3   clinical characteristics: Adrenalectomy for ACA   \n",
       "4               tumor diameter (cm): not applicable   \n",
       "5                 tumor weight (gm): not applicable   \n",
       "6              weiss score of tumor: not applicable   \n",
       "7             mitotic rate of tumor: not applicable   \n",
       "8                       tumor stage: not applicable   \n",
       "9                   years to last followup: unknown   \n",
       "10          dead or alive at last followup: unknown   \n",
       "11                  beta-catenin staining: Membrane   \n",
       "12            beta-catenin sequence result: unknown   \n",
       "13                      hormone expression: unknown   \n",
       "\n",
       "                                            GSM825372  \\\n",
       "0                                             age: 28   \n",
       "1                                              Sex: M   \n",
       "2                                  side of body: Left   \n",
       "3   clinical characteristics: Adrenalectomy for ph...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                    beta-catenin staining: Membrane   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                            GSM825373  \\\n",
       "0                                             age: 55   \n",
       "1                                              Sex: M   \n",
       "2                                  side of body: Left   \n",
       "3   clinical characteristics: Adrenalectomy for mu...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                    beta-catenin staining: Membrane   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                            GSM825374  \\\n",
       "0                                             age: 78   \n",
       "1                                              Sex: M   \n",
       "2                                  side of body: Left   \n",
       "3   clinical characteristics: Adrenalectomy for re...   \n",
       "4                 tumor diameter (cm): not applicable   \n",
       "5                   tumor weight (gm): not applicable   \n",
       "6                weiss score of tumor: not applicable   \n",
       "7               mitotic rate of tumor: not applicable   \n",
       "8                         tumor stage: not applicable   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                     beta-catenin staining: unknown   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                        hormone expression: unknown   \n",
       "\n",
       "                                            GSM825375  ...  \\\n",
       "0                                             age: 41  ...   \n",
       "1                                              Sex: M  ...   \n",
       "2                               side of body: unknown  ...   \n",
       "3   clinical characteristics: Adrenalectomy for re...  ...   \n",
       "4                 tumor diameter (cm): not applicable  ...   \n",
       "5                   tumor weight (gm): not applicable  ...   \n",
       "6                weiss score of tumor: not applicable  ...   \n",
       "7               mitotic rate of tumor: not applicable  ...   \n",
       "8                         tumor stage: not applicable  ...   \n",
       "9                     years to last followup: unknown  ...   \n",
       "10            dead or alive at last followup: unknown  ...   \n",
       "11                     beta-catenin staining: unknown  ...   \n",
       "12              beta-catenin sequence result: unknown  ...   \n",
       "13                        hormone expression: unknown  ...   \n",
       "\n",
       "                                      GSM825422  \\\n",
       "0                                       age: 19   \n",
       "1                                        Sex: M   \n",
       "2                            side of body: Left   \n",
       "3   clinical characteristics: Left adrenal mass   \n",
       "4                     tumor diameter (cm): 10.5   \n",
       "5                    tumor weight (gm): unknown   \n",
       "6                    weiss score of tumor: High   \n",
       "7                     mitotic rate of tumor: 27   \n",
       "8                                tumor stage: 2   \n",
       "9           years to last followup: 0.619178082   \n",
       "10         dead or alive at last followup: dead   \n",
       "11               beta-catenin staining: Nuclear   \n",
       "12        beta-catenin sequence result: unknown   \n",
       "13                 hormone expression: Cortisol   \n",
       "\n",
       "                                     GSM825423  \\\n",
       "0                                      age: 61   \n",
       "1                                       Sex: F   \n",
       "2                          side of body: Right   \n",
       "3   clinical characteristics: Cushing Syndrome   \n",
       "4                     tumor diameter (cm): 9.8   \n",
       "5                       tumor weight (gm): 253   \n",
       "6                   weiss score of tumor: High   \n",
       "7                    mitotic rate of tumor: 27   \n",
       "8                               tumor stage: 2   \n",
       "9          years to last followup: 0.942465753   \n",
       "10        dead or alive at last followup: dead   \n",
       "11              beta-catenin staining: Nuclear   \n",
       "12       beta-catenin sequence result: unknown   \n",
       "13                hormone expression: Cortisol   \n",
       "\n",
       "                                            GSM825424  \\\n",
       "0                                             age: 33   \n",
       "1                                              Sex: F   \n",
       "2                                  side of body: Left   \n",
       "3   clinical characteristics: Cortisol secreting t...   \n",
       "4                             tumor diameter (cm): 13   \n",
       "5                              tumor weight (gm): 660   \n",
       "6                          weiss score of tumor: High   \n",
       "7                           mitotic rate of tumor: 70   \n",
       "8                                      tumor stage: 3   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                     beta-catenin staining: Nuclear   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                       hormone expression: Cortisol   \n",
       "\n",
       "                                            GSM825425  \\\n",
       "0                                             age: 31   \n",
       "1                                              Sex: F   \n",
       "2                        side of body: not applicable   \n",
       "3   clinical characteristics: History of Adrenocor...   \n",
       "4                        tumor diameter (cm): unknown   \n",
       "5                          tumor weight (gm): unknown   \n",
       "6                           weiss score of tumor: Low   \n",
       "7                           mitotic rate of tumor: 16   \n",
       "8                                      tumor stage: 4   \n",
       "9                     years to last followup: unknown   \n",
       "10            dead or alive at last followup: unknown   \n",
       "11                     beta-catenin staining: Nuclear   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                 hormone expression: No_information   \n",
       "\n",
       "                                     GSM825426  \\\n",
       "0                                      age: 60   \n",
       "1                                       Sex: F   \n",
       "2                          side of body: Right   \n",
       "3   clinical characteristics: Cushing Syndrome   \n",
       "4                    tumor diameter (cm): 18.5   \n",
       "5                      tumor weight (gm): 1805   \n",
       "6                   weiss score of tumor: High   \n",
       "7                    mitotic rate of tumor: 40   \n",
       "8                               tumor stage: 2   \n",
       "9          years to last followup: 0.364383562   \n",
       "10        dead or alive at last followup: dead   \n",
       "11              beta-catenin staining: Nuclear   \n",
       "12       beta-catenin sequence result: unknown   \n",
       "13                hormone expression: Cortisol   \n",
       "\n",
       "                                            GSM825427  \\\n",
       "0                                             age: 51   \n",
       "1                                              Sex: F   \n",
       "2                                 side of body: Right   \n",
       "3   clinical characteristics: Right adrenal mass; ...   \n",
       "4                           tumor diameter (cm): 15.5   \n",
       "5                             tumor weight (gm): 1305   \n",
       "6                          weiss score of tumor: High   \n",
       "7                           mitotic rate of tumor: 23   \n",
       "8                                      tumor stage: 3   \n",
       "9                 years to last followup: 2.232876712   \n",
       "10               dead or alive at last followup: dead   \n",
       "11                     beta-catenin staining: Nuclear   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                   hormone expression: Testosterone   \n",
       "\n",
       "                                            GSM825428  \\\n",
       "0                                             age: 71   \n",
       "1                                              Sex: M   \n",
       "2                                 side of body: Right   \n",
       "3   clinical characteristics: Right adrenal tumor;...   \n",
       "4                             tumor diameter (cm): 22   \n",
       "5                             tumor weight (gm): 2440   \n",
       "6                          weiss score of tumor: High   \n",
       "7                           mitotic rate of tumor: 22   \n",
       "8                                      tumor stage: 4   \n",
       "9                 years to last followup: 0.624657534   \n",
       "10               dead or alive at last followup: dead   \n",
       "11                     beta-catenin staining: Nuclear   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                 hormone expression: Non-functional   \n",
       "\n",
       "                                            GSM825429  \\\n",
       "0                                             age: 25   \n",
       "1                                              Sex: F   \n",
       "2                                 side of body: Right   \n",
       "3   clinical characteristics: History of adrenal m...   \n",
       "4                             tumor diameter (cm): 26   \n",
       "5                             tumor weight (gm): 2300   \n",
       "6                          weiss score of tumor: High   \n",
       "7                           mitotic rate of tumor: 34   \n",
       "8                                      tumor stage: 4   \n",
       "9                 years to last followup: 0.898630137   \n",
       "10               dead or alive at last followup: dead   \n",
       "11                     beta-catenin staining: Nuclear   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13          hormone expression: Cortisol+Testosterone   \n",
       "\n",
       "                                            GSM825430  \\\n",
       "0                                             age: 47   \n",
       "1                                              Sex: F   \n",
       "2                                 side of body: Right   \n",
       "3   clinical characteristics: Cushing Syndrome; Ri...   \n",
       "4                        tumor diameter (cm): unknown   \n",
       "5                          tumor weight (gm): unknown   \n",
       "6                          weiss score of tumor: High   \n",
       "7                           mitotic rate of tumor: 37   \n",
       "8                                      tumor stage: 1   \n",
       "9                        years to last followup: 7.81   \n",
       "10              dead or alive at last followup: alive   \n",
       "11                     beta-catenin staining: Nuclear   \n",
       "12              beta-catenin sequence result: unknown   \n",
       "13                       hormone expression: Cortisol   \n",
       "\n",
       "                                            GSM825431  \n",
       "0                                             age: 49  \n",
       "1                                              Sex: F  \n",
       "2                                  side of body: Left  \n",
       "3   clinical characteristics: Elevated serum aldos...  \n",
       "4                            tumor diameter (cm): 8.7  \n",
       "5                              tumor weight (gm): 397  \n",
       "6                          weiss score of tumor: High  \n",
       "7                           mitotic rate of tumor: 28  \n",
       "8                                      tumor stage: 2  \n",
       "9                     years to last followup: unknown  \n",
       "10            dead or alive at last followup: unknown  \n",
       "11                     beta-catenin staining: unknown  \n",
       "12              beta-catenin sequence result: unknown  \n",
       "13                 hormone expression: No_information  \n",
       "\n",
       "[14 rows x 66 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE33371\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'weiss score of tumor: not applicable',\n",
       "       'weiss score of tumor: High', 'weiss score of tumor: Low'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[6]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6\n",
       "weiss score of tumor: not applicable    32\n",
       "weiss score of tumor: High              20\n",
       "weiss score of tumor: Low               13\n",
       "!Sample_characteristics_ch1              1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_counts = tumor_stage_row.value_counts()\n",
    "tumor_stage_counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM825367</th>\n",
       "      <th>GSM825368</th>\n",
       "      <th>GSM825369</th>\n",
       "      <th>GSM825370</th>\n",
       "      <th>GSM825371</th>\n",
       "      <th>GSM825372</th>\n",
       "      <th>GSM825373</th>\n",
       "      <th>GSM825374</th>\n",
       "      <th>GSM825375</th>\n",
       "      <th>GSM825376</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM825422</th>\n",
       "      <th>GSM825423</th>\n",
       "      <th>GSM825424</th>\n",
       "      <th>GSM825425</th>\n",
       "      <th>GSM825426</th>\n",
       "      <th>GSM825427</th>\n",
       "      <th>GSM825428</th>\n",
       "      <th>GSM825429</th>\n",
       "      <th>GSM825430</th>\n",
       "      <th>GSM825431</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>3.426349</td>\n",
       "      <td>3.507046</td>\n",
       "      <td>3.503791</td>\n",
       "      <td>3.483730</td>\n",
       "      <td>3.593175</td>\n",
       "      <td>3.388279</td>\n",
       "      <td>3.323458</td>\n",
       "      <td>3.274850</td>\n",
       "      <td>3.367542</td>\n",
       "      <td>3.458184</td>\n",
       "      <td>...</td>\n",
       "      <td>3.629919</td>\n",
       "      <td>3.613207</td>\n",
       "      <td>3.721563</td>\n",
       "      <td>3.629715</td>\n",
       "      <td>3.336460</td>\n",
       "      <td>3.820530</td>\n",
       "      <td>4.179609</td>\n",
       "      <td>3.808751</td>\n",
       "      <td>3.198107</td>\n",
       "      <td>4.115178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>2.707570</td>\n",
       "      <td>2.615950</td>\n",
       "      <td>2.390935</td>\n",
       "      <td>2.568202</td>\n",
       "      <td>2.609594</td>\n",
       "      <td>2.567026</td>\n",
       "      <td>2.592177</td>\n",
       "      <td>2.635484</td>\n",
       "      <td>2.485721</td>\n",
       "      <td>2.726727</td>\n",
       "      <td>...</td>\n",
       "      <td>2.745855</td>\n",
       "      <td>2.640481</td>\n",
       "      <td>2.729165</td>\n",
       "      <td>2.755875</td>\n",
       "      <td>2.434569</td>\n",
       "      <td>2.678518</td>\n",
       "      <td>2.472756</td>\n",
       "      <td>2.545307</td>\n",
       "      <td>2.694605</td>\n",
       "      <td>3.105851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>2.725912</td>\n",
       "      <td>2.958564</td>\n",
       "      <td>2.517196</td>\n",
       "      <td>2.913284</td>\n",
       "      <td>2.880814</td>\n",
       "      <td>2.670246</td>\n",
       "      <td>2.710963</td>\n",
       "      <td>2.809560</td>\n",
       "      <td>2.525045</td>\n",
       "      <td>2.656098</td>\n",
       "      <td>...</td>\n",
       "      <td>2.322219</td>\n",
       "      <td>2.328380</td>\n",
       "      <td>2.334454</td>\n",
       "      <td>2.361728</td>\n",
       "      <td>2.260071</td>\n",
       "      <td>3.437909</td>\n",
       "      <td>2.158362</td>\n",
       "      <td>2.457882</td>\n",
       "      <td>2.444045</td>\n",
       "      <td>2.264818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>2.986772</td>\n",
       "      <td>3.145196</td>\n",
       "      <td>3.185542</td>\n",
       "      <td>3.296665</td>\n",
       "      <td>3.298416</td>\n",
       "      <td>3.145196</td>\n",
       "      <td>3.084219</td>\n",
       "      <td>3.203577</td>\n",
       "      <td>3.289589</td>\n",
       "      <td>3.124178</td>\n",
       "      <td>...</td>\n",
       "      <td>2.883093</td>\n",
       "      <td>3.007321</td>\n",
       "      <td>2.942008</td>\n",
       "      <td>2.873321</td>\n",
       "      <td>2.885361</td>\n",
       "      <td>2.829947</td>\n",
       "      <td>2.801404</td>\n",
       "      <td>3.012837</td>\n",
       "      <td>3.133219</td>\n",
       "      <td>2.977266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>2.214844</td>\n",
       "      <td>2.155336</td>\n",
       "      <td>2.178977</td>\n",
       "      <td>2.198657</td>\n",
       "      <td>2.004321</td>\n",
       "      <td>2.167317</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>2.053078</td>\n",
       "      <td>2.176091</td>\n",
       "      <td>2.232996</td>\n",
       "      <td>...</td>\n",
       "      <td>2.117271</td>\n",
       "      <td>2.278754</td>\n",
       "      <td>2.190332</td>\n",
       "      <td>2.158362</td>\n",
       "      <td>2.167317</td>\n",
       "      <td>2.127105</td>\n",
       "      <td>1.973128</td>\n",
       "      <td>2.220108</td>\n",
       "      <td>2.315970</td>\n",
       "      <td>2.107210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>2.060698</td>\n",
       "      <td>1.977724</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>2.060698</td>\n",
       "      <td>2.086360</td>\n",
       "      <td>2.093422</td>\n",
       "      <td>2.152288</td>\n",
       "      <td>2.089905</td>\n",
       "      <td>1.954243</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>...</td>\n",
       "      <td>2.071882</td>\n",
       "      <td>2.075547</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.167317</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.071882</td>\n",
       "      <td>2.021189</td>\n",
       "      <td>2.075547</td>\n",
       "      <td>1.924279</td>\n",
       "      <td>2.045323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>1.949390</td>\n",
       "      <td>2.012837</td>\n",
       "      <td>2.130334</td>\n",
       "      <td>2.187521</td>\n",
       "      <td>2.082785</td>\n",
       "      <td>2.056905</td>\n",
       "      <td>2.017033</td>\n",
       "      <td>2.117271</td>\n",
       "      <td>2.110590</td>\n",
       "      <td>2.089905</td>\n",
       "      <td>...</td>\n",
       "      <td>2.206826</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.117271</td>\n",
       "      <td>2.152288</td>\n",
       "      <td>2.149219</td>\n",
       "      <td>2.123852</td>\n",
       "      <td>2.086360</td>\n",
       "      <td>2.096910</td>\n",
       "      <td>2.178977</td>\n",
       "      <td>2.037426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>2.086360</td>\n",
       "      <td>2.060698</td>\n",
       "      <td>2.025306</td>\n",
       "      <td>2.012837</td>\n",
       "      <td>1.982271</td>\n",
       "      <td>2.089905</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>...</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>2.082785</td>\n",
       "      <td>2.075547</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>2.025306</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.060698</td>\n",
       "      <td>2.017033</td>\n",
       "      <td>2.025306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>2.060698</td>\n",
       "      <td>2.107210</td>\n",
       "      <td>1.934498</td>\n",
       "      <td>2.056905</td>\n",
       "      <td>2.110590</td>\n",
       "      <td>2.146128</td>\n",
       "      <td>2.107210</td>\n",
       "      <td>2.103804</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>...</td>\n",
       "      <td>2.146128</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>1.982271</td>\n",
       "      <td>2.021189</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>2.195900</td>\n",
       "      <td>2.045323</td>\n",
       "      <td>2.060698</td>\n",
       "      <td>2.107210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.120574</td>\n",
       "      <td>2.056905</td>\n",
       "      <td>2.152288</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.143015</td>\n",
       "      <td>1.995635</td>\n",
       "      <td>2.113943</td>\n",
       "      <td>1.968483</td>\n",
       "      <td>2.021189</td>\n",
       "      <td>...</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.045323</td>\n",
       "      <td>1.954243</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>1.982271</td>\n",
       "      <td>1.954243</td>\n",
       "      <td>1.903090</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>1.968483</td>\n",
       "      <td>2.064458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM825367  GSM825368  GSM825369  GSM825370  GSM825371  \\\n",
       "ID                                                                       \n",
       "1007_s_at         3.426349   3.507046   3.503791   3.483730   3.593175   \n",
       "1053_at           2.707570   2.615950   2.390935   2.568202   2.609594   \n",
       "117_at            2.725912   2.958564   2.517196   2.913284   2.880814   \n",
       "121_at            2.986772   3.145196   3.185542   3.296665   3.298416   \n",
       "1255_g_at         2.214844   2.155336   2.178977   2.198657   2.004321   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "AFFX-ThrX-5_at    2.060698   1.977724   2.008600   2.060698   2.086360   \n",
       "AFFX-ThrX-M_at    1.949390   2.012837   2.130334   2.187521   2.082785   \n",
       "AFFX-TrpnX-3_at   2.139879   2.041393   2.086360   2.060698   2.025306   \n",
       "AFFX-TrpnX-5_at   2.060698   2.107210   1.934498   2.056905   2.110590   \n",
       "AFFX-TrpnX-M_at   2.049218   2.120574   2.056905   2.152288   2.033424   \n",
       "\n",
       "                 GSM825372  GSM825373  GSM825374  GSM825375  GSM825376  ...  \\\n",
       "ID                                                                      ...   \n",
       "1007_s_at         3.388279   3.323458   3.274850   3.367542   3.458184  ...   \n",
       "1053_at           2.567026   2.592177   2.635484   2.485721   2.726727  ...   \n",
       "117_at            2.670246   2.710963   2.809560   2.525045   2.656098  ...   \n",
       "121_at            3.145196   3.084219   3.203577   3.289589   3.124178  ...   \n",
       "1255_g_at         2.167317   2.079181   2.053078   2.176091   2.232996  ...   \n",
       "...                    ...        ...        ...        ...        ...  ...   \n",
       "AFFX-ThrX-5_at    2.093422   2.152288   2.089905   1.954243   2.079181  ...   \n",
       "AFFX-ThrX-M_at    2.056905   2.017033   2.117271   2.110590   2.089905  ...   \n",
       "AFFX-TrpnX-3_at   2.012837   1.982271   2.089905   2.008600   2.041393  ...   \n",
       "AFFX-TrpnX-5_at   2.146128   2.107210   2.103804   2.079181   2.041393  ...   \n",
       "AFFX-TrpnX-M_at   2.143015   1.995635   2.113943   1.968483   2.021189  ...   \n",
       "\n",
       "                 GSM825422  GSM825423  GSM825424  GSM825425  GSM825426  \\\n",
       "ID                                                                       \n",
       "1007_s_at         3.629919   3.613207   3.721563   3.629715   3.336460   \n",
       "1053_at           2.745855   2.640481   2.729165   2.755875   2.434569   \n",
       "117_at            2.322219   2.328380   2.334454   2.361728   2.260071   \n",
       "121_at            2.883093   3.007321   2.942008   2.873321   2.885361   \n",
       "1255_g_at         2.117271   2.278754   2.190332   2.158362   2.167317   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "AFFX-ThrX-5_at    2.071882   2.075547   2.049218   2.167317   2.033424   \n",
       "AFFX-ThrX-M_at    2.206826   2.049218   2.117271   2.152288   2.149219   \n",
       "AFFX-TrpnX-3_at   2.008600   2.041393   2.082785   2.075547   2.041393   \n",
       "AFFX-TrpnX-5_at   2.146128   2.041393   1.982271   2.021189   2.049218   \n",
       "AFFX-TrpnX-M_at   2.033424   2.045323   1.954243   2.079181   1.982271   \n",
       "\n",
       "                 GSM825427  GSM825428  GSM825429  GSM825430  GSM825431  \n",
       "ID                                                                      \n",
       "1007_s_at         3.820530   4.179609   3.808751   3.198107   4.115178  \n",
       "1053_at           2.678518   2.472756   2.545307   2.694605   3.105851  \n",
       "117_at            3.437909   2.158362   2.457882   2.444045   2.264818  \n",
       "121_at            2.829947   2.801404   3.012837   3.133219   2.977266  \n",
       "1255_g_at         2.127105   1.973128   2.220108   2.315970   2.107210  \n",
       "...                    ...        ...        ...        ...        ...  \n",
       "AFFX-ThrX-5_at    2.071882   2.021189   2.075547   1.924279   2.045323  \n",
       "AFFX-ThrX-M_at    2.123852   2.086360   2.096910   2.178977   2.037426  \n",
       "AFFX-TrpnX-3_at   2.025306   2.049218   2.060698   2.017033   2.025306  \n",
       "AFFX-TrpnX-5_at   2.008600   2.195900   2.045323   2.060698   2.107210  \n",
       "AFFX-TrpnX-M_at   1.954243   1.903090   2.008600   1.968483   2.064458  \n",
       "\n",
       "[54675 rows x 65 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 6\n",
    "age_row = 0\n",
    "gender_row = 1\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'weiss score of tumor: High':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM825367</th>\n",
       "      <th>GSM825368</th>\n",
       "      <th>GSM825369</th>\n",
       "      <th>GSM825370</th>\n",
       "      <th>GSM825371</th>\n",
       "      <th>GSM825372</th>\n",
       "      <th>GSM825373</th>\n",
       "      <th>GSM825374</th>\n",
       "      <th>GSM825375</th>\n",
       "      <th>GSM825376</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM825422</th>\n",
       "      <th>GSM825423</th>\n",
       "      <th>GSM825424</th>\n",
       "      <th>GSM825425</th>\n",
       "      <th>GSM825426</th>\n",
       "      <th>GSM825427</th>\n",
       "      <th>GSM825428</th>\n",
       "      <th>GSM825429</th>\n",
       "      <th>GSM825430</th>\n",
       "      <th>GSM825431</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>71</td>\n",
       "      <td>58</td>\n",
       "      <td>71</td>\n",
       "      <td>44</td>\n",
       "      <td>32</td>\n",
       "      <td>28</td>\n",
       "      <td>55</td>\n",
       "      <td>78</td>\n",
       "      <td>41</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>61</td>\n",
       "      <td>33</td>\n",
       "      <td>31</td>\n",
       "      <td>60</td>\n",
       "      <td>51</td>\n",
       "      <td>71</td>\n",
       "      <td>25</td>\n",
       "      <td>47</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM825367  GSM825368  GSM825369  GSM825370  GSM825371  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "Age                           71         58         71         44         32   \n",
       "Gender                         1          1          0          0          1   \n",
       "\n",
       "                       GSM825372  GSM825373  GSM825374  GSM825375  GSM825376  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "Age                           28         55         78         41         58   \n",
       "Gender                         0          0          0          0          0   \n",
       "\n",
       "                       ...  GSM825422  GSM825423  GSM825424  GSM825425  \\\n",
       "Adrenocortical Cancer  ...          1          1          1          0   \n",
       "Age                    ...         19         61         33         31   \n",
       "Gender                 ...          0          1          1          1   \n",
       "\n",
       "                       GSM825426  GSM825427  GSM825428  GSM825429  GSM825430  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "Age                           60         51         71         25         47   \n",
       "Gender                         1          1          0          1          1   \n",
       "\n",
       "                       GSM825431  \n",
       "Adrenocortical Cancer          1  \n",
       "Age                           49  \n",
       "Gender                         1  \n",
       "\n",
       "[3 rows x 65 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608610</th>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>2.045322979</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608611</th>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>2.037426498</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608612</th>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>2.025305865</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608613</th>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>2.10720997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608614</th>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>2.064457989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3608615 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID       GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0              1007_s_at       U48705     NaN            Homo sapiens   \n",
       "1                1053_at       M87338     NaN            Homo sapiens   \n",
       "2                 117_at       X51757     NaN            Homo sapiens   \n",
       "3                 121_at       X69699     NaN            Homo sapiens   \n",
       "4              1255_g_at       L36861     NaN            Homo sapiens   \n",
       "...                  ...          ...     ...                     ...   \n",
       "3608610   AFFX-ThrX-5_at  2.045322979     NaN                     NaN   \n",
       "3608611   AFFX-ThrX-M_at  2.037426498     NaN                     NaN   \n",
       "3608612  AFFX-TrpnX-3_at  2.025305865     NaN                     NaN   \n",
       "3608613  AFFX-TrpnX-5_at   2.10720997     NaN                     NaN   \n",
       "3608614  AFFX-TrpnX-M_at  2.064457989     NaN                     NaN   \n",
       "\n",
       "        Annotation Date      Sequence Type                  Sequence Source  \\\n",
       "0           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "2           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "4           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                 ...                ...                              ...   \n",
       "3608610             NaN                NaN                              NaN   \n",
       "3608611             NaN                NaN                              NaN   \n",
       "3608612             NaN                NaN                              NaN   \n",
       "3608613             NaN                NaN                              NaN   \n",
       "3608614             NaN                NaN                              NaN   \n",
       "\n",
       "                                        Target Description  \\\n",
       "0        U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1        M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2        X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3        X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4        L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                    ...   \n",
       "3608610                                                NaN   \n",
       "3608611                                                NaN   \n",
       "3608612                                                NaN   \n",
       "3608613                                                NaN   \n",
       "3608614                                                NaN   \n",
       "\n",
       "        Representative Public ID  \\\n",
       "0                         U48705   \n",
       "1                         M87338   \n",
       "2                         X51757   \n",
       "3                         X69699   \n",
       "4                         L36861   \n",
       "...                          ...   \n",
       "3608610                      NaN   \n",
       "3608611                      NaN   \n",
       "3608612                      NaN   \n",
       "3608613                      NaN   \n",
       "3608614                      NaN   \n",
       "\n",
       "                                                Gene Title       Gene Symbol  \\\n",
       "0        discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1              replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2                     heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                             paired box 8              PAX8   \n",
       "4                  guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "...                                                    ...               ...   \n",
       "3608610                                                NaN               NaN   \n",
       "3608611                                                NaN               NaN   \n",
       "3608612                                                NaN               NaN   \n",
       "3608613                                                NaN               NaN   \n",
       "3608614                                                NaN               NaN   \n",
       "\n",
       "            ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0        780 /// 100616237  NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1                     5982  NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2                     3310                                          NM_002155   \n",
       "3                     7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                     2978                         NM_000409 /// XM_006715073   \n",
       "...                    ...                                                ...   \n",
       "3608610                NaN                                                NaN   \n",
       "3608611                NaN                                                NaN   \n",
       "3608612                NaN                                                NaN   \n",
       "3608613                NaN                                                NaN   \n",
       "3608614                NaN                                                NaN   \n",
       "\n",
       "                          Gene Ontology Biological Process  \\\n",
       "0        0001558 // regulation of cell growth // inferr...   \n",
       "1        0000278 // mitotic cell cycle // traceable aut...   \n",
       "2        0000902 // cell morphogenesis // inferred from...   \n",
       "3        0001655 // urogenital system development // in...   \n",
       "4        0007165 // signal transduction // non-traceabl...   \n",
       "...                                                    ...   \n",
       "3608610                                                NaN   \n",
       "3608611                                                NaN   \n",
       "3608612                                                NaN   \n",
       "3608613                                                NaN   \n",
       "3608614                                                NaN   \n",
       "\n",
       "                          Gene Ontology Cellular Component  \\\n",
       "0        0005576 // extracellular region // inferred fr...   \n",
       "1        0005634 // nucleus // inferred from electronic...   \n",
       "2        0005737 // cytoplasm // inferred from direct a...   \n",
       "3        0005634 // nucleus // inferred from direct ass...   \n",
       "4        0001750 // photoreceptor outer segment // infe...   \n",
       "...                                                    ...   \n",
       "3608610                                                NaN   \n",
       "3608611                                                NaN   \n",
       "3608612                                                NaN   \n",
       "3608613                                                NaN   \n",
       "3608614                                                NaN   \n",
       "\n",
       "                          Gene Ontology Molecular Function  \n",
       "0        0000166 // nucleotide binding // inferred from...  \n",
       "1        0000166 // nucleotide binding // inferred from...  \n",
       "2        0000166 // nucleotide binding // inferred from...  \n",
       "3        0000979 // RNA polymerase II core promoter seq...  \n",
       "4        0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                    ...  \n",
       "3608610                                                NaN  \n",
       "3608611                                                NaN  \n",
       "3608612                                                NaN  \n",
       "3608613                                                NaN  \n",
       "3608614                                                NaN  \n",
       "\n",
       "[3608615 rows x 16 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM825367</th>\n",
       "      <th>GSM825368</th>\n",
       "      <th>GSM825369</th>\n",
       "      <th>GSM825370</th>\n",
       "      <th>GSM825371</th>\n",
       "      <th>GSM825372</th>\n",
       "      <th>GSM825373</th>\n",
       "      <th>GSM825374</th>\n",
       "      <th>GSM825375</th>\n",
       "      <th>GSM825376</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM825422</th>\n",
       "      <th>GSM825423</th>\n",
       "      <th>GSM825424</th>\n",
       "      <th>GSM825425</th>\n",
       "      <th>GSM825426</th>\n",
       "      <th>GSM825427</th>\n",
       "      <th>GSM825428</th>\n",
       "      <th>GSM825429</th>\n",
       "      <th>GSM825430</th>\n",
       "      <th>GSM825431</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>4.124145</td>\n",
       "      <td>3.991625</td>\n",
       "      <td>4.013174</td>\n",
       "      <td>4.125416</td>\n",
       "      <td>4.197391</td>\n",
       "      <td>4.258757</td>\n",
       "      <td>4.234467</td>\n",
       "      <td>4.217115</td>\n",
       "      <td>4.161817</td>\n",
       "      <td>4.077186</td>\n",
       "      <td>...</td>\n",
       "      <td>4.035990</td>\n",
       "      <td>4.203848</td>\n",
       "      <td>3.621592</td>\n",
       "      <td>4.276737</td>\n",
       "      <td>3.664736</td>\n",
       "      <td>3.584218</td>\n",
       "      <td>4.253338</td>\n",
       "      <td>3.346744</td>\n",
       "      <td>3.927165</td>\n",
       "      <td>3.870755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>3.019947</td>\n",
       "      <td>2.647383</td>\n",
       "      <td>3.001734</td>\n",
       "      <td>3.035830</td>\n",
       "      <td>3.135769</td>\n",
       "      <td>2.778151</td>\n",
       "      <td>2.825426</td>\n",
       "      <td>2.909021</td>\n",
       "      <td>2.846955</td>\n",
       "      <td>2.879096</td>\n",
       "      <td>...</td>\n",
       "      <td>2.804821</td>\n",
       "      <td>3.066699</td>\n",
       "      <td>2.849419</td>\n",
       "      <td>3.006894</td>\n",
       "      <td>2.369216</td>\n",
       "      <td>2.510545</td>\n",
       "      <td>2.975432</td>\n",
       "      <td>2.702431</td>\n",
       "      <td>2.469822</td>\n",
       "      <td>3.063333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>3.019947</td>\n",
       "      <td>2.647383</td>\n",
       "      <td>3.001734</td>\n",
       "      <td>3.035830</td>\n",
       "      <td>3.135769</td>\n",
       "      <td>2.778151</td>\n",
       "      <td>2.825426</td>\n",
       "      <td>2.909021</td>\n",
       "      <td>2.846955</td>\n",
       "      <td>2.879096</td>\n",
       "      <td>...</td>\n",
       "      <td>2.804821</td>\n",
       "      <td>3.066699</td>\n",
       "      <td>2.849419</td>\n",
       "      <td>3.006894</td>\n",
       "      <td>2.369216</td>\n",
       "      <td>2.510545</td>\n",
       "      <td>2.975432</td>\n",
       "      <td>2.702431</td>\n",
       "      <td>2.469822</td>\n",
       "      <td>3.063333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1P2</th>\n",
       "      <td>1.913814</td>\n",
       "      <td>2.100371</td>\n",
       "      <td>1.732394</td>\n",
       "      <td>2.143015</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>1.903090</td>\n",
       "      <td>2.143015</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>1.968483</td>\n",
       "      <td>1.869232</td>\n",
       "      <td>...</td>\n",
       "      <td>1.995635</td>\n",
       "      <td>1.913814</td>\n",
       "      <td>2.075547</td>\n",
       "      <td>2.110590</td>\n",
       "      <td>2.103804</td>\n",
       "      <td>1.944483</td>\n",
       "      <td>2.176091</td>\n",
       "      <td>1.991226</td>\n",
       "      <td>2.133539</td>\n",
       "      <td>1.838849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC078883.4</th>\n",
       "      <td>2.025306</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>2.222716</td>\n",
       "      <td>2.283301</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.037426</td>\n",
       "      <td>2.158362</td>\n",
       "      <td>2.267172</td>\n",
       "      <td>1.968483</td>\n",
       "      <td>2.075547</td>\n",
       "      <td>...</td>\n",
       "      <td>1.698970</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.176091</td>\n",
       "      <td>1.963788</td>\n",
       "      <td>2.238046</td>\n",
       "      <td>2.238046</td>\n",
       "      <td>2.143015</td>\n",
       "      <td>2.045323</td>\n",
       "      <td>2.146128</td>\n",
       "      <td>2.139879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abParts</th>\n",
       "      <td>2.164353</td>\n",
       "      <td>2.235528</td>\n",
       "      <td>1.875061</td>\n",
       "      <td>2.117271</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>1.982271</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>2.303196</td>\n",
       "      <td>2.082785</td>\n",
       "      <td>2.064458</td>\n",
       "      <td>...</td>\n",
       "      <td>2.025306</td>\n",
       "      <td>2.086360</td>\n",
       "      <td>2.056905</td>\n",
       "      <td>2.071882</td>\n",
       "      <td>1.954243</td>\n",
       "      <td>2.143015</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>2.645422</td>\n",
       "      <td>1.982271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>2.201397</td>\n",
       "      <td>2.071882</td>\n",
       "      <td>2.190332</td>\n",
       "      <td>2.217484</td>\n",
       "      <td>2.217484</td>\n",
       "      <td>2.296665</td>\n",
       "      <td>2.152288</td>\n",
       "      <td>2.257679</td>\n",
       "      <td>2.146128</td>\n",
       "      <td>2.217484</td>\n",
       "      <td>...</td>\n",
       "      <td>2.149219</td>\n",
       "      <td>2.127105</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.190332</td>\n",
       "      <td>2.250420</td>\n",
       "      <td>2.198657</td>\n",
       "      <td>2.195900</td>\n",
       "      <td>2.123852</td>\n",
       "      <td>2.363612</td>\n",
       "      <td>2.206826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>av27s1</th>\n",
       "      <td>2.071882</td>\n",
       "      <td>2.264818</td>\n",
       "      <td>2.181844</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.220108</td>\n",
       "      <td>2.214844</td>\n",
       "      <td>2.089905</td>\n",
       "      <td>2.071882</td>\n",
       "      <td>2.113943</td>\n",
       "      <td>2.155336</td>\n",
       "      <td>...</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.012837</td>\n",
       "      <td>2.230449</td>\n",
       "      <td>2.240549</td>\n",
       "      <td>2.127105</td>\n",
       "      <td>2.113943</td>\n",
       "      <td>2.209515</td>\n",
       "      <td>1.812913</td>\n",
       "      <td>2.041393</td>\n",
       "      <td>2.089905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>2.146748</td>\n",
       "      <td>2.133586</td>\n",
       "      <td>2.044953</td>\n",
       "      <td>2.289606</td>\n",
       "      <td>1.982977</td>\n",
       "      <td>2.155803</td>\n",
       "      <td>2.258677</td>\n",
       "      <td>2.145195</td>\n",
       "      <td>2.027077</td>\n",
       "      <td>1.972587</td>\n",
       "      <td>...</td>\n",
       "      <td>2.293249</td>\n",
       "      <td>2.185691</td>\n",
       "      <td>2.087756</td>\n",
       "      <td>2.199196</td>\n",
       "      <td>2.367256</td>\n",
       "      <td>2.628767</td>\n",
       "      <td>2.489699</td>\n",
       "      <td>2.148333</td>\n",
       "      <td>2.150298</td>\n",
       "      <td>2.277354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mir-223</th>\n",
       "      <td>2.612784</td>\n",
       "      <td>2.481443</td>\n",
       "      <td>2.535294</td>\n",
       "      <td>2.532754</td>\n",
       "      <td>2.445604</td>\n",
       "      <td>2.274158</td>\n",
       "      <td>2.376577</td>\n",
       "      <td>2.356026</td>\n",
       "      <td>2.324282</td>\n",
       "      <td>2.260071</td>\n",
       "      <td>...</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.053078</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.008600</td>\n",
       "      <td>1.973128</td>\n",
       "      <td>2.107210</td>\n",
       "      <td>1.973128</td>\n",
       "      <td>2.230449</td>\n",
       "      <td>2.456366</td>\n",
       "      <td>2.064458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25851 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GSM825367  GSM825368  GSM825369  GSM825370  GSM825371  \\\n",
       "Gene                                                                   \n",
       " ABCB4          4.124145   3.991625   4.013174   4.125416   4.197391   \n",
       " ABCC6P1        3.019947   2.647383   3.001734   3.035830   3.135769   \n",
       " ABCC6P2        3.019947   2.647383   3.001734   3.035830   3.135769   \n",
       " ABCD1P2        1.913814   2.100371   1.732394   2.143015   2.041393   \n",
       " AC078883.4     2.025306   2.008600   2.222716   2.283301   2.139879   \n",
       "...                  ...        ...        ...        ...        ...   \n",
       "abParts         2.164353   2.235528   1.875061   2.117271   2.033424   \n",
       "alpha           2.201397   2.071882   2.190332   2.217484   2.217484   \n",
       "av27s1          2.071882   2.264818   2.181844   2.033424   2.220108   \n",
       "hsa-let-7a-3    2.146748   2.133586   2.044953   2.289606   1.982977   \n",
       "mir-223         2.612784   2.481443   2.535294   2.532754   2.445604   \n",
       "\n",
       "               GSM825372  GSM825373  GSM825374  GSM825375  GSM825376  ...  \\\n",
       "Gene                                                                  ...   \n",
       " ABCB4          4.258757   4.234467   4.217115   4.161817   4.077186  ...   \n",
       " ABCC6P1        2.778151   2.825426   2.909021   2.846955   2.879096  ...   \n",
       " ABCC6P2        2.778151   2.825426   2.909021   2.846955   2.879096  ...   \n",
       " ABCD1P2        1.903090   2.143015   2.049218   1.968483   1.869232  ...   \n",
       " AC078883.4     2.037426   2.158362   2.267172   1.968483   2.075547  ...   \n",
       "...                  ...        ...        ...        ...        ...  ...   \n",
       "abParts         1.982271   2.008600   2.303196   2.082785   2.064458  ...   \n",
       "alpha           2.296665   2.152288   2.257679   2.146128   2.217484  ...   \n",
       "av27s1          2.214844   2.089905   2.071882   2.113943   2.155336  ...   \n",
       "hsa-let-7a-3    2.155803   2.258677   2.145195   2.027077   1.972587  ...   \n",
       "mir-223         2.274158   2.376577   2.356026   2.324282   2.260071  ...   \n",
       "\n",
       "               GSM825422  GSM825423  GSM825424  GSM825425  GSM825426  \\\n",
       "Gene                                                                   \n",
       " ABCB4          4.035990   4.203848   3.621592   4.276737   3.664736   \n",
       " ABCC6P1        2.804821   3.066699   2.849419   3.006894   2.369216   \n",
       " ABCC6P2        2.804821   3.066699   2.849419   3.006894   2.369216   \n",
       " ABCD1P2        1.995635   1.913814   2.075547   2.110590   2.103804   \n",
       " AC078883.4     1.698970   2.000000   2.176091   1.963788   2.238046   \n",
       "...                  ...        ...        ...        ...        ...   \n",
       "abParts         2.025306   2.086360   2.056905   2.071882   1.954243   \n",
       "alpha           2.149219   2.127105   2.139879   2.190332   2.250420   \n",
       "av27s1          2.139879   2.012837   2.230449   2.240549   2.127105   \n",
       "hsa-let-7a-3    2.293249   2.185691   2.087756   2.199196   2.367256   \n",
       "mir-223         2.033424   2.053078   2.049218   2.008600   1.973128   \n",
       "\n",
       "               GSM825427  GSM825428  GSM825429  GSM825430  GSM825431  \n",
       "Gene                                                                  \n",
       " ABCB4          3.584218   4.253338   3.346744   3.927165   3.870755  \n",
       " ABCC6P1        2.510545   2.975432   2.702431   2.469822   3.063333  \n",
       " ABCC6P2        2.510545   2.975432   2.702431   2.469822   3.063333  \n",
       " ABCD1P2        1.944483   2.176091   1.991226   2.133539   1.838849  \n",
       " AC078883.4     2.238046   2.143015   2.045323   2.146128   2.139879  \n",
       "...                  ...        ...        ...        ...        ...  \n",
       "abParts         2.143015   2.033424   2.079181   2.645422   1.982271  \n",
       "alpha           2.198657   2.195900   2.123852   2.363612   2.206826  \n",
       "av27s1          2.113943   2.209515   1.812913   2.041393   2.089905  \n",
       "hsa-let-7a-3    2.628767   2.489699   2.148333   2.150298   2.277354  \n",
       "mir-223         2.107210   1.973128   2.230449   2.456366   2.064458  \n",
       "\n",
       "[25851 rows x 65 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM825367</th>\n",
       "      <th>GSM825368</th>\n",
       "      <th>GSM825369</th>\n",
       "      <th>GSM825370</th>\n",
       "      <th>GSM825371</th>\n",
       "      <th>GSM825372</th>\n",
       "      <th>GSM825373</th>\n",
       "      <th>GSM825374</th>\n",
       "      <th>GSM825375</th>\n",
       "      <th>GSM825376</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM825422</th>\n",
       "      <th>GSM825423</th>\n",
       "      <th>GSM825424</th>\n",
       "      <th>GSM825425</th>\n",
       "      <th>GSM825426</th>\n",
       "      <th>GSM825427</th>\n",
       "      <th>GSM825428</th>\n",
       "      <th>GSM825429</th>\n",
       "      <th>GSM825430</th>\n",
       "      <th>GSM825431</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>2.344392</td>\n",
       "      <td>2.274158</td>\n",
       "      <td>2.409933</td>\n",
       "      <td>2.220108</td>\n",
       "      <td>2.382017</td>\n",
       "      <td>2.230449</td>\n",
       "      <td>2.225309</td>\n",
       "      <td>2.235528</td>\n",
       "      <td>2.453318</td>\n",
       "      <td>2.220108</td>\n",
       "      <td>...</td>\n",
       "      <td>2.770852</td>\n",
       "      <td>2.190332</td>\n",
       "      <td>2.492760</td>\n",
       "      <td>2.729165</td>\n",
       "      <td>2.176091</td>\n",
       "      <td>2.465383</td>\n",
       "      <td>2.100371</td>\n",
       "      <td>2.195900</td>\n",
       "      <td>2.841985</td>\n",
       "      <td>2.120574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>2.225309</td>\n",
       "      <td>2.152288</td>\n",
       "      <td>2.029384</td>\n",
       "      <td>2.209515</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.029384</td>\n",
       "      <td>1.949390</td>\n",
       "      <td>1.838849</td>\n",
       "      <td>2.068186</td>\n",
       "      <td>...</td>\n",
       "      <td>2.214844</td>\n",
       "      <td>2.071882</td>\n",
       "      <td>2.127105</td>\n",
       "      <td>2.158362</td>\n",
       "      <td>2.113943</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.012837</td>\n",
       "      <td>2.086360</td>\n",
       "      <td>1.838849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>2.258387</td>\n",
       "      <td>2.215618</td>\n",
       "      <td>2.275090</td>\n",
       "      <td>2.361580</td>\n",
       "      <td>2.336395</td>\n",
       "      <td>2.325736</td>\n",
       "      <td>2.125941</td>\n",
       "      <td>2.275871</td>\n",
       "      <td>2.282481</td>\n",
       "      <td>2.282451</td>\n",
       "      <td>...</td>\n",
       "      <td>2.159084</td>\n",
       "      <td>2.174736</td>\n",
       "      <td>2.309599</td>\n",
       "      <td>2.302383</td>\n",
       "      <td>2.312800</td>\n",
       "      <td>2.161667</td>\n",
       "      <td>2.235580</td>\n",
       "      <td>2.187923</td>\n",
       "      <td>2.321805</td>\n",
       "      <td>2.199543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>3.241844</td>\n",
       "      <td>3.274122</td>\n",
       "      <td>3.237545</td>\n",
       "      <td>3.226205</td>\n",
       "      <td>3.260617</td>\n",
       "      <td>3.224270</td>\n",
       "      <td>3.173422</td>\n",
       "      <td>3.203154</td>\n",
       "      <td>3.231411</td>\n",
       "      <td>3.281136</td>\n",
       "      <td>...</td>\n",
       "      <td>3.021480</td>\n",
       "      <td>2.958363</td>\n",
       "      <td>3.048116</td>\n",
       "      <td>2.932255</td>\n",
       "      <td>2.986461</td>\n",
       "      <td>3.246799</td>\n",
       "      <td>2.960222</td>\n",
       "      <td>3.199634</td>\n",
       "      <td>3.262587</td>\n",
       "      <td>3.331146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>2.758155</td>\n",
       "      <td>2.844477</td>\n",
       "      <td>2.843855</td>\n",
       "      <td>2.719331</td>\n",
       "      <td>2.836324</td>\n",
       "      <td>2.894870</td>\n",
       "      <td>2.625312</td>\n",
       "      <td>2.561101</td>\n",
       "      <td>2.684845</td>\n",
       "      <td>2.831870</td>\n",
       "      <td>...</td>\n",
       "      <td>2.799341</td>\n",
       "      <td>2.161368</td>\n",
       "      <td>2.703291</td>\n",
       "      <td>2.773786</td>\n",
       "      <td>3.175512</td>\n",
       "      <td>2.797268</td>\n",
       "      <td>2.649335</td>\n",
       "      <td>2.850646</td>\n",
       "      <td>2.676694</td>\n",
       "      <td>2.894870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>2.082785</td>\n",
       "      <td>2.222716</td>\n",
       "      <td>2.206826</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.428135</td>\n",
       "      <td>2.267172</td>\n",
       "      <td>2.133539</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.193125</td>\n",
       "      <td>2.285557</td>\n",
       "      <td>...</td>\n",
       "      <td>2.190332</td>\n",
       "      <td>2.332438</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>2.056905</td>\n",
       "      <td>2.064458</td>\n",
       "      <td>2.195900</td>\n",
       "      <td>2.037426</td>\n",
       "      <td>2.225309</td>\n",
       "      <td>1.995635</td>\n",
       "      <td>2.167317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>3.290175</td>\n",
       "      <td>3.154617</td>\n",
       "      <td>3.074636</td>\n",
       "      <td>3.071638</td>\n",
       "      <td>3.251223</td>\n",
       "      <td>3.233694</td>\n",
       "      <td>3.158933</td>\n",
       "      <td>3.164545</td>\n",
       "      <td>3.286878</td>\n",
       "      <td>3.168897</td>\n",
       "      <td>...</td>\n",
       "      <td>3.123692</td>\n",
       "      <td>3.093839</td>\n",
       "      <td>3.288377</td>\n",
       "      <td>3.194621</td>\n",
       "      <td>3.121960</td>\n",
       "      <td>3.163507</td>\n",
       "      <td>3.073203</td>\n",
       "      <td>3.172625</td>\n",
       "      <td>3.054001</td>\n",
       "      <td>3.049166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>3.073090</td>\n",
       "      <td>3.153078</td>\n",
       "      <td>3.055331</td>\n",
       "      <td>3.152671</td>\n",
       "      <td>3.171448</td>\n",
       "      <td>3.329625</td>\n",
       "      <td>3.251829</td>\n",
       "      <td>3.499841</td>\n",
       "      <td>2.973627</td>\n",
       "      <td>3.033582</td>\n",
       "      <td>...</td>\n",
       "      <td>2.835737</td>\n",
       "      <td>2.869494</td>\n",
       "      <td>2.826289</td>\n",
       "      <td>2.711672</td>\n",
       "      <td>3.043238</td>\n",
       "      <td>3.207796</td>\n",
       "      <td>2.891020</td>\n",
       "      <td>2.983929</td>\n",
       "      <td>3.430660</td>\n",
       "      <td>3.085020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>2.613386</td>\n",
       "      <td>2.654351</td>\n",
       "      <td>2.645732</td>\n",
       "      <td>2.545531</td>\n",
       "      <td>2.657407</td>\n",
       "      <td>2.639701</td>\n",
       "      <td>2.723974</td>\n",
       "      <td>2.662660</td>\n",
       "      <td>2.614772</td>\n",
       "      <td>2.551009</td>\n",
       "      <td>...</td>\n",
       "      <td>3.047227</td>\n",
       "      <td>2.924887</td>\n",
       "      <td>2.622894</td>\n",
       "      <td>3.019029</td>\n",
       "      <td>2.529522</td>\n",
       "      <td>2.599417</td>\n",
       "      <td>2.869427</td>\n",
       "      <td>2.567629</td>\n",
       "      <td>2.702477</td>\n",
       "      <td>2.543037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>3.053991</td>\n",
       "      <td>3.050952</td>\n",
       "      <td>2.958490</td>\n",
       "      <td>3.008351</td>\n",
       "      <td>3.107854</td>\n",
       "      <td>3.163569</td>\n",
       "      <td>3.011254</td>\n",
       "      <td>2.929139</td>\n",
       "      <td>2.974821</td>\n",
       "      <td>2.989448</td>\n",
       "      <td>...</td>\n",
       "      <td>2.928948</td>\n",
       "      <td>3.089858</td>\n",
       "      <td>3.183842</td>\n",
       "      <td>3.164391</td>\n",
       "      <td>3.165522</td>\n",
       "      <td>3.029347</td>\n",
       "      <td>2.954086</td>\n",
       "      <td>3.177093</td>\n",
       "      <td>2.988465</td>\n",
       "      <td>2.836813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM825367  GSM825368  GSM825369  GSM825370  GSM825371  GSM825372  \\\n",
       "A1BG       2.344392   2.274158   2.409933   2.220108   2.382017   2.230449   \n",
       "A1BG-AS1   2.225309   2.152288   2.029384   2.209515   2.033424   2.033424   \n",
       "A1CF       2.258387   2.215618   2.275090   2.361580   2.336395   2.325736   \n",
       "A2M        3.241844   3.274122   3.237545   3.226205   3.260617   3.224270   \n",
       "A2M-AS1    2.758155   2.844477   2.843855   2.719331   2.836324   2.894870   \n",
       "...             ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A     2.082785   2.222716   2.206826   2.139879   2.428135   2.267172   \n",
       "ZYG11B     3.290175   3.154617   3.074636   3.071638   3.251223   3.233694   \n",
       "ZYX        3.073090   3.153078   3.055331   3.152671   3.171448   3.329625   \n",
       "ZZEF1      2.613386   2.654351   2.645732   2.545531   2.657407   2.639701   \n",
       "ZZZ3       3.053991   3.050952   2.958490   3.008351   3.107854   3.163569   \n",
       "\n",
       "          GSM825373  GSM825374  GSM825375  GSM825376  ...  GSM825422  \\\n",
       "A1BG       2.225309   2.235528   2.453318   2.220108  ...   2.770852   \n",
       "A1BG-AS1   2.029384   1.949390   1.838849   2.068186  ...   2.214844   \n",
       "A1CF       2.125941   2.275871   2.282481   2.282451  ...   2.159084   \n",
       "A2M        3.173422   3.203154   3.231411   3.281136  ...   3.021480   \n",
       "A2M-AS1    2.625312   2.561101   2.684845   2.831870  ...   2.799341   \n",
       "...             ...        ...        ...        ...  ...        ...   \n",
       "ZYG11A     2.133539   2.139879   2.193125   2.285557  ...   2.190332   \n",
       "ZYG11B     3.158933   3.164545   3.286878   3.168897  ...   3.123692   \n",
       "ZYX        3.251829   3.499841   2.973627   3.033582  ...   2.835737   \n",
       "ZZEF1      2.723974   2.662660   2.614772   2.551009  ...   3.047227   \n",
       "ZZZ3       3.011254   2.929139   2.974821   2.989448  ...   2.928948   \n",
       "\n",
       "          GSM825423  GSM825424  GSM825425  GSM825426  GSM825427  GSM825428  \\\n",
       "A1BG       2.190332   2.492760   2.729165   2.176091   2.465383   2.100371   \n",
       "A1BG-AS1   2.071882   2.127105   2.158362   2.113943   2.079181   2.049218   \n",
       "A1CF       2.174736   2.309599   2.302383   2.312800   2.161667   2.235580   \n",
       "A2M        2.958363   3.048116   2.932255   2.986461   3.246799   2.960222   \n",
       "A2M-AS1    2.161368   2.703291   2.773786   3.175512   2.797268   2.649335   \n",
       "...             ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A     2.332438   2.139879   2.056905   2.064458   2.195900   2.037426   \n",
       "ZYG11B     3.093839   3.288377   3.194621   3.121960   3.163507   3.073203   \n",
       "ZYX        2.869494   2.826289   2.711672   3.043238   3.207796   2.891020   \n",
       "ZZEF1      2.924887   2.622894   3.019029   2.529522   2.599417   2.869427   \n",
       "ZZZ3       3.089858   3.183842   3.164391   3.165522   3.029347   2.954086   \n",
       "\n",
       "          GSM825429  GSM825430  GSM825431  \n",
       "A1BG       2.195900   2.841985   2.120574  \n",
       "A1BG-AS1   2.012837   2.086360   1.838849  \n",
       "A1CF       2.187923   2.321805   2.199543  \n",
       "A2M        3.199634   3.262587   3.331146  \n",
       "A2M-AS1    2.850646   2.676694   2.894870  \n",
       "...             ...        ...        ...  \n",
       "ZYG11A     2.225309   1.995635   2.167317  \n",
       "ZYG11B     3.172625   3.054001   3.049166  \n",
       "ZYX        2.983929   3.430660   3.085020  \n",
       "ZZEF1      2.567629   2.702477   2.543037  \n",
       "ZZZ3       3.177093   2.988465   2.836813  \n",
       "\n",
       "[17857 rows x 65 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM825367</th>\n",
       "      <td>0.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.344392</td>\n",
       "      <td>2.225309</td>\n",
       "      <td>2.258387</td>\n",
       "      <td>3.241844</td>\n",
       "      <td>2.758155</td>\n",
       "      <td>2.177419</td>\n",
       "      <td>2.376577</td>\n",
       "      <td>...</td>\n",
       "      <td>2.531471</td>\n",
       "      <td>2.750508</td>\n",
       "      <td>2.662758</td>\n",
       "      <td>2.467995</td>\n",
       "      <td>2.540189</td>\n",
       "      <td>2.082785</td>\n",
       "      <td>3.290175</td>\n",
       "      <td>3.07309</td>\n",
       "      <td>2.613386</td>\n",
       "      <td>3.053991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825368</th>\n",
       "      <td>0.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.274158</td>\n",
       "      <td>2.152288</td>\n",
       "      <td>2.215618</td>\n",
       "      <td>3.274122</td>\n",
       "      <td>2.844477</td>\n",
       "      <td>2.216885</td>\n",
       "      <td>2.287802</td>\n",
       "      <td>...</td>\n",
       "      <td>2.5887</td>\n",
       "      <td>2.812913</td>\n",
       "      <td>2.716838</td>\n",
       "      <td>2.359753</td>\n",
       "      <td>2.554861</td>\n",
       "      <td>2.222716</td>\n",
       "      <td>3.154617</td>\n",
       "      <td>3.153078</td>\n",
       "      <td>2.654351</td>\n",
       "      <td>3.050952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825369</th>\n",
       "      <td>0.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.409933</td>\n",
       "      <td>2.029384</td>\n",
       "      <td>2.27509</td>\n",
       "      <td>3.237545</td>\n",
       "      <td>2.843855</td>\n",
       "      <td>2.306074</td>\n",
       "      <td>2.515874</td>\n",
       "      <td>...</td>\n",
       "      <td>2.500122</td>\n",
       "      <td>2.900913</td>\n",
       "      <td>2.682145</td>\n",
       "      <td>2.441456</td>\n",
       "      <td>2.402199</td>\n",
       "      <td>2.206826</td>\n",
       "      <td>3.074636</td>\n",
       "      <td>3.055331</td>\n",
       "      <td>2.645732</td>\n",
       "      <td>2.95849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825370</th>\n",
       "      <td>0.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.220108</td>\n",
       "      <td>2.209515</td>\n",
       "      <td>2.36158</td>\n",
       "      <td>3.226205</td>\n",
       "      <td>2.719331</td>\n",
       "      <td>2.101516</td>\n",
       "      <td>2.350248</td>\n",
       "      <td>...</td>\n",
       "      <td>2.601973</td>\n",
       "      <td>2.93852</td>\n",
       "      <td>2.691081</td>\n",
       "      <td>2.340014</td>\n",
       "      <td>2.449002</td>\n",
       "      <td>2.139879</td>\n",
       "      <td>3.071638</td>\n",
       "      <td>3.152671</td>\n",
       "      <td>2.545531</td>\n",
       "      <td>3.008351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825371</th>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.382017</td>\n",
       "      <td>2.033424</td>\n",
       "      <td>2.336395</td>\n",
       "      <td>3.260617</td>\n",
       "      <td>2.836324</td>\n",
       "      <td>2.11548</td>\n",
       "      <td>2.525045</td>\n",
       "      <td>...</td>\n",
       "      <td>2.582165</td>\n",
       "      <td>2.814913</td>\n",
       "      <td>2.699838</td>\n",
       "      <td>2.388292</td>\n",
       "      <td>2.532311</td>\n",
       "      <td>2.428135</td>\n",
       "      <td>3.251223</td>\n",
       "      <td>3.171448</td>\n",
       "      <td>2.657407</td>\n",
       "      <td>3.107854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825427</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.465383</td>\n",
       "      <td>2.079181</td>\n",
       "      <td>2.161667</td>\n",
       "      <td>3.246799</td>\n",
       "      <td>2.797268</td>\n",
       "      <td>2.164251</td>\n",
       "      <td>2.146128</td>\n",
       "      <td>...</td>\n",
       "      <td>2.739914</td>\n",
       "      <td>3.301247</td>\n",
       "      <td>2.647383</td>\n",
       "      <td>2.446042</td>\n",
       "      <td>2.946991</td>\n",
       "      <td>2.1959</td>\n",
       "      <td>3.163507</td>\n",
       "      <td>3.207796</td>\n",
       "      <td>2.599417</td>\n",
       "      <td>3.029347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825428</th>\n",
       "      <td>1.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.100371</td>\n",
       "      <td>2.049218</td>\n",
       "      <td>2.23558</td>\n",
       "      <td>2.960222</td>\n",
       "      <td>2.649335</td>\n",
       "      <td>2.120561</td>\n",
       "      <td>2.217484</td>\n",
       "      <td>...</td>\n",
       "      <td>3.011802</td>\n",
       "      <td>3.423082</td>\n",
       "      <td>2.350248</td>\n",
       "      <td>2.336012</td>\n",
       "      <td>2.730562</td>\n",
       "      <td>2.037426</td>\n",
       "      <td>3.073203</td>\n",
       "      <td>2.89102</td>\n",
       "      <td>2.869427</td>\n",
       "      <td>2.954086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825429</th>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.1959</td>\n",
       "      <td>2.012837</td>\n",
       "      <td>2.187923</td>\n",
       "      <td>3.199634</td>\n",
       "      <td>2.850646</td>\n",
       "      <td>2.136501</td>\n",
       "      <td>2.32838</td>\n",
       "      <td>...</td>\n",
       "      <td>2.911548</td>\n",
       "      <td>3.344981</td>\n",
       "      <td>2.40654</td>\n",
       "      <td>2.345784</td>\n",
       "      <td>2.62128</td>\n",
       "      <td>2.225309</td>\n",
       "      <td>3.172625</td>\n",
       "      <td>2.983929</td>\n",
       "      <td>2.567629</td>\n",
       "      <td>3.177093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825430</th>\n",
       "      <td>1.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.841985</td>\n",
       "      <td>2.08636</td>\n",
       "      <td>2.321805</td>\n",
       "      <td>3.262587</td>\n",
       "      <td>2.676694</td>\n",
       "      <td>2.119374</td>\n",
       "      <td>2.20412</td>\n",
       "      <td>...</td>\n",
       "      <td>2.940211</td>\n",
       "      <td>3.49304</td>\n",
       "      <td>2.506505</td>\n",
       "      <td>2.330053</td>\n",
       "      <td>2.492831</td>\n",
       "      <td>1.995635</td>\n",
       "      <td>3.054001</td>\n",
       "      <td>3.43066</td>\n",
       "      <td>2.702477</td>\n",
       "      <td>2.988465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM825431</th>\n",
       "      <td>1.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.120574</td>\n",
       "      <td>1.838849</td>\n",
       "      <td>2.199543</td>\n",
       "      <td>3.331146</td>\n",
       "      <td>2.89487</td>\n",
       "      <td>2.170638</td>\n",
       "      <td>2.31597</td>\n",
       "      <td>...</td>\n",
       "      <td>3.226104</td>\n",
       "      <td>3.936313</td>\n",
       "      <td>2.674861</td>\n",
       "      <td>2.526635</td>\n",
       "      <td>2.820016</td>\n",
       "      <td>2.167317</td>\n",
       "      <td>3.049166</td>\n",
       "      <td>3.08502</td>\n",
       "      <td>2.543037</td>\n",
       "      <td>2.836813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 17860 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Adrenocortical Cancer   Age Gender      A1BG  A1BG-AS1      A1CF  \\\n",
       "GSM825367                   0.0  71.0    1.0  2.344392  2.225309  2.258387   \n",
       "GSM825368                   0.0  58.0    1.0  2.274158  2.152288  2.215618   \n",
       "GSM825369                   0.0  71.0    0.0  2.409933  2.029384   2.27509   \n",
       "GSM825370                   0.0  44.0    0.0  2.220108  2.209515   2.36158   \n",
       "GSM825371                   0.0  32.0    1.0  2.382017  2.033424  2.336395   \n",
       "...                         ...   ...    ...       ...       ...       ...   \n",
       "GSM825427                   1.0  51.0    1.0  2.465383  2.079181  2.161667   \n",
       "GSM825428                   1.0  71.0    0.0  2.100371  2.049218   2.23558   \n",
       "GSM825429                   1.0  25.0    1.0    2.1959  2.012837  2.187923   \n",
       "GSM825430                   1.0  47.0    1.0  2.841985   2.08636  2.321805   \n",
       "GSM825431                   1.0  49.0    1.0  2.120574  1.838849  2.199543   \n",
       "\n",
       "                A2M   A2M-AS1     A2ML1     A2MP1  ...    ZWILCH     ZWINT  \\\n",
       "GSM825367  3.241844  2.758155  2.177419  2.376577  ...  2.531471  2.750508   \n",
       "GSM825368  3.274122  2.844477  2.216885  2.287802  ...    2.5887  2.812913   \n",
       "GSM825369  3.237545  2.843855  2.306074  2.515874  ...  2.500122  2.900913   \n",
       "GSM825370  3.226205  2.719331  2.101516  2.350248  ...  2.601973   2.93852   \n",
       "GSM825371  3.260617  2.836324   2.11548  2.525045  ...  2.582165  2.814913   \n",
       "...             ...       ...       ...       ...  ...       ...       ...   \n",
       "GSM825427  3.246799  2.797268  2.164251  2.146128  ...  2.739914  3.301247   \n",
       "GSM825428  2.960222  2.649335  2.120561  2.217484  ...  3.011802  3.423082   \n",
       "GSM825429  3.199634  2.850646  2.136501   2.32838  ...  2.911548  3.344981   \n",
       "GSM825430  3.262587  2.676694  2.119374   2.20412  ...  2.940211   3.49304   \n",
       "GSM825431  3.331146   2.89487  2.170638   2.31597  ...  3.226104  3.936313   \n",
       "\n",
       "               ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX  \\\n",
       "GSM825367  2.662758  2.467995  2.540189  2.082785  3.290175   3.07309   \n",
       "GSM825368  2.716838  2.359753  2.554861  2.222716  3.154617  3.153078   \n",
       "GSM825369  2.682145  2.441456  2.402199  2.206826  3.074636  3.055331   \n",
       "GSM825370  2.691081  2.340014  2.449002  2.139879  3.071638  3.152671   \n",
       "GSM825371  2.699838  2.388292  2.532311  2.428135  3.251223  3.171448   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "GSM825427  2.647383  2.446042  2.946991    2.1959  3.163507  3.207796   \n",
       "GSM825428  2.350248  2.336012  2.730562  2.037426  3.073203   2.89102   \n",
       "GSM825429   2.40654  2.345784   2.62128  2.225309  3.172625  2.983929   \n",
       "GSM825430  2.506505  2.330053  2.492831  1.995635  3.054001   3.43066   \n",
       "GSM825431  2.674861  2.526635  2.820016  2.167317  3.049166   3.08502   \n",
       "\n",
       "              ZZEF1      ZZZ3  \n",
       "GSM825367  2.613386  3.053991  \n",
       "GSM825368  2.654351  3.050952  \n",
       "GSM825369  2.645732   2.95849  \n",
       "GSM825370  2.545531  3.008351  \n",
       "GSM825371  2.657407  3.107854  \n",
       "...             ...       ...  \n",
       "GSM825427  2.599417  3.029347  \n",
       "GSM825428  2.869427  2.954086  \n",
       "GSM825429  2.567629  3.177093  \n",
       "GSM825430  2.702477  2.988465  \n",
       "GSM825431  2.543037  2.836813  \n",
       "\n",
       "[63 rows x 17860 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 63 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 18 occurrences. This represents 28.57% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 38.5\n",
      "  50% (Median): 48.0\n",
      "  75%: 56.0\n",
      "Min: 19.0\n",
      "Max: 87.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 23 occurrences. This represents 36.51% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Integrated genomic analyses of adrenocortical tumors (SNP array, DNA methylation, mRNA and miRNA expression).\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1196390</th>\n",
       "      <th>GSM1196391</th>\n",
       "      <th>GSM1196392</th>\n",
       "      <th>GSM1196393</th>\n",
       "      <th>GSM1196394</th>\n",
       "      <th>GSM1196395</th>\n",
       "      <th>GSM1196396</th>\n",
       "      <th>GSM1196397</th>\n",
       "      <th>GSM1196398</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196418</th>\n",
       "      <th>GSM1196419</th>\n",
       "      <th>GSM1196420</th>\n",
       "      <th>GSM1196421</th>\n",
       "      <th>GSM1196422</th>\n",
       "      <th>GSM1196423</th>\n",
       "      <th>GSM1196424</th>\n",
       "      <th>GSM1196425</th>\n",
       "      <th>GSM1196426</th>\n",
       "      <th>GSM1196427</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM1196390  \\\n",
       "0  !Sample_characteristics_ch1  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196391                           GSM1196392  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196393                           GSM1196394  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196395                           GSM1196396  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196397                           GSM1196398  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "   ...                           GSM1196418  \\\n",
       "0  ...  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196419                           GSM1196420  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196421                           GSM1196422  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196423                           GSM1196424  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196425                           GSM1196426  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196427  \n",
       "0  cell type: Adrenocortical carcinoma  \n",
       "\n",
       "[1 rows x 39 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# error\n",
    "cohort = accession_num = \"GSE49280\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'cell type: Adrenocortical carcinoma'], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196390</th>\n",
       "      <th>GSM1196391</th>\n",
       "      <th>GSM1196392</th>\n",
       "      <th>GSM1196393</th>\n",
       "      <th>GSM1196394</th>\n",
       "      <th>GSM1196395</th>\n",
       "      <th>GSM1196396</th>\n",
       "      <th>GSM1196397</th>\n",
       "      <th>GSM1196398</th>\n",
       "      <th>GSM1196399</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196418</th>\n",
       "      <th>GSM1196419</th>\n",
       "      <th>GSM1196420</th>\n",
       "      <th>GSM1196421</th>\n",
       "      <th>GSM1196422</th>\n",
       "      <th>GSM1196423</th>\n",
       "      <th>GSM1196424</th>\n",
       "      <th>GSM1196425</th>\n",
       "      <th>GSM1196426</th>\n",
       "      <th>GSM1196427</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs1000000</th>\n",
       "      <td>0.1065</td>\n",
       "      <td>0.2021</td>\n",
       "      <td>-0.0258</td>\n",
       "      <td>0.0714</td>\n",
       "      <td>0.1821</td>\n",
       "      <td>0.3716</td>\n",
       "      <td>-0.0343</td>\n",
       "      <td>0.0632</td>\n",
       "      <td>0.4660</td>\n",
       "      <td>0.1609</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0428</td>\n",
       "      <td>-0.0564</td>\n",
       "      <td>0.0930</td>\n",
       "      <td>0.1396</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0.0441</td>\n",
       "      <td>0.0616</td>\n",
       "      <td>0.1482</td>\n",
       "      <td>0.0390</td>\n",
       "      <td>0.0873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000002</th>\n",
       "      <td>0.1184</td>\n",
       "      <td>-0.2697</td>\n",
       "      <td>-0.5366</td>\n",
       "      <td>-0.3392</td>\n",
       "      <td>0.1135</td>\n",
       "      <td>-0.5014</td>\n",
       "      <td>-0.2013</td>\n",
       "      <td>-0.3576</td>\n",
       "      <td>0.0717</td>\n",
       "      <td>-0.1686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1097</td>\n",
       "      <td>0.1616</td>\n",
       "      <td>-0.2828</td>\n",
       "      <td>-0.2522</td>\n",
       "      <td>0.2335</td>\n",
       "      <td>-0.1791</td>\n",
       "      <td>0.2011</td>\n",
       "      <td>0.2515</td>\n",
       "      <td>-0.5280</td>\n",
       "      <td>-0.1804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs10000023</th>\n",
       "      <td>-0.0627</td>\n",
       "      <td>0.0830</td>\n",
       "      <td>-0.1168</td>\n",
       "      <td>-0.0207</td>\n",
       "      <td>-0.0500</td>\n",
       "      <td>0.1033</td>\n",
       "      <td>-0.0169</td>\n",
       "      <td>0.0197</td>\n",
       "      <td>0.2101</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1189</td>\n",
       "      <td>0.0611</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>-0.0478</td>\n",
       "      <td>0.1226</td>\n",
       "      <td>-0.2908</td>\n",
       "      <td>0.1440</td>\n",
       "      <td>0.0704</td>\n",
       "      <td>-0.0951</td>\n",
       "      <td>-0.1550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000003</th>\n",
       "      <td>-0.0874</td>\n",
       "      <td>-0.5206</td>\n",
       "      <td>-0.2224</td>\n",
       "      <td>-0.1689</td>\n",
       "      <td>-0.0666</td>\n",
       "      <td>-0.2013</td>\n",
       "      <td>-0.1730</td>\n",
       "      <td>0.1010</td>\n",
       "      <td>0.2544</td>\n",
       "      <td>-0.1112</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0022</td>\n",
       "      <td>0.0930</td>\n",
       "      <td>-0.1605</td>\n",
       "      <td>-0.1667</td>\n",
       "      <td>0.0316</td>\n",
       "      <td>0.0493</td>\n",
       "      <td>-0.1910</td>\n",
       "      <td>0.0574</td>\n",
       "      <td>-0.1625</td>\n",
       "      <td>-0.1756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs10000030</th>\n",
       "      <td>0.7767</td>\n",
       "      <td>-0.0285</td>\n",
       "      <td>0.4749</td>\n",
       "      <td>0.2115</td>\n",
       "      <td>-0.0876</td>\n",
       "      <td>-0.3540</td>\n",
       "      <td>0.4442</td>\n",
       "      <td>-0.5963</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>-0.4139</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3487</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.5774</td>\n",
       "      <td>0.4109</td>\n",
       "      <td>0.3663</td>\n",
       "      <td>0.2121</td>\n",
       "      <td>0.6078</td>\n",
       "      <td>0.8196</td>\n",
       "      <td>0.2831</td>\n",
       "      <td>0.8315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34742</th>\n",
       "      <td>0.1940</td>\n",
       "      <td>0.2780</td>\n",
       "      <td>0.2586</td>\n",
       "      <td>-0.1372</td>\n",
       "      <td>0.3958</td>\n",
       "      <td>0.1807</td>\n",
       "      <td>0.2403</td>\n",
       "      <td>0.1901</td>\n",
       "      <td>0.3416</td>\n",
       "      <td>0.4133</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2485</td>\n",
       "      <td>0.0686</td>\n",
       "      <td>0.1226</td>\n",
       "      <td>-0.0354</td>\n",
       "      <td>-0.1239</td>\n",
       "      <td>-0.3011</td>\n",
       "      <td>0.0032</td>\n",
       "      <td>-0.0218</td>\n",
       "      <td>-0.4481</td>\n",
       "      <td>0.0698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34743</th>\n",
       "      <td>0.2213</td>\n",
       "      <td>0.3447</td>\n",
       "      <td>0.2913</td>\n",
       "      <td>-0.2419</td>\n",
       "      <td>0.3323</td>\n",
       "      <td>0.1770</td>\n",
       "      <td>0.2181</td>\n",
       "      <td>0.0668</td>\n",
       "      <td>0.3055</td>\n",
       "      <td>0.4727</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2333</td>\n",
       "      <td>0.2030</td>\n",
       "      <td>0.1948</td>\n",
       "      <td>0.0367</td>\n",
       "      <td>0.0499</td>\n",
       "      <td>-0.1797</td>\n",
       "      <td>0.1574</td>\n",
       "      <td>-0.0151</td>\n",
       "      <td>-0.3437</td>\n",
       "      <td>0.1108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34744</th>\n",
       "      <td>-0.1274</td>\n",
       "      <td>-0.0201</td>\n",
       "      <td>-0.0016</td>\n",
       "      <td>-0.4076</td>\n",
       "      <td>0.2235</td>\n",
       "      <td>-0.0196</td>\n",
       "      <td>-0.0525</td>\n",
       "      <td>-0.0919</td>\n",
       "      <td>0.2553</td>\n",
       "      <td>0.2095</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0968</td>\n",
       "      <td>-0.0961</td>\n",
       "      <td>-0.1108</td>\n",
       "      <td>-0.4564</td>\n",
       "      <td>-0.4473</td>\n",
       "      <td>-0.7310</td>\n",
       "      <td>-0.1037</td>\n",
       "      <td>-0.5226</td>\n",
       "      <td>-0.7002</td>\n",
       "      <td>-0.2425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34761</th>\n",
       "      <td>-0.1124</td>\n",
       "      <td>0.0917</td>\n",
       "      <td>0.0726</td>\n",
       "      <td>-0.3928</td>\n",
       "      <td>0.2356</td>\n",
       "      <td>0.0827</td>\n",
       "      <td>-0.0679</td>\n",
       "      <td>0.1435</td>\n",
       "      <td>0.3957</td>\n",
       "      <td>0.2965</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2568</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0917</td>\n",
       "      <td>-0.1877</td>\n",
       "      <td>-0.1994</td>\n",
       "      <td>-0.4965</td>\n",
       "      <td>-0.0968</td>\n",
       "      <td>-0.2921</td>\n",
       "      <td>-0.5193</td>\n",
       "      <td>-0.0352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS35706</th>\n",
       "      <td>0.0560</td>\n",
       "      <td>0.2881</td>\n",
       "      <td>0.2553</td>\n",
       "      <td>-0.3665</td>\n",
       "      <td>0.3393</td>\n",
       "      <td>-0.2476</td>\n",
       "      <td>0.1102</td>\n",
       "      <td>-0.2108</td>\n",
       "      <td>0.3091</td>\n",
       "      <td>0.3059</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5293</td>\n",
       "      <td>-0.5513</td>\n",
       "      <td>-0.4255</td>\n",
       "      <td>-0.5476</td>\n",
       "      <td>-1.1190</td>\n",
       "      <td>-1.4563</td>\n",
       "      <td>-0.9340</td>\n",
       "      <td>-0.8940</td>\n",
       "      <td>-1.8872</td>\n",
       "      <td>-0.3105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>710272 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM1196390  GSM1196391  GSM1196392  GSM1196393  GSM1196394  \\\n",
       "ID                                                                       \n",
       "rs1000000       0.1065      0.2021     -0.0258      0.0714      0.1821   \n",
       "rs1000002       0.1184     -0.2697     -0.5366     -0.3392      0.1135   \n",
       "rs10000023     -0.0627      0.0830     -0.1168     -0.0207     -0.0500   \n",
       "rs1000003      -0.0874     -0.5206     -0.2224     -0.1689     -0.0666   \n",
       "rs10000030      0.7767     -0.0285      0.4749      0.2115     -0.0876   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "VGXS34742       0.1940      0.2780      0.2586     -0.1372      0.3958   \n",
       "VGXS34743       0.2213      0.3447      0.2913     -0.2419      0.3323   \n",
       "VGXS34744      -0.1274     -0.0201     -0.0016     -0.4076      0.2235   \n",
       "VGXS34761      -0.1124      0.0917      0.0726     -0.3928      0.2356   \n",
       "VGXS35706       0.0560      0.2881      0.2553     -0.3665      0.3393   \n",
       "\n",
       "            GSM1196395  GSM1196396  GSM1196397  GSM1196398  GSM1196399  ...  \\\n",
       "ID                                                                      ...   \n",
       "rs1000000       0.3716     -0.0343      0.0632      0.4660      0.1609  ...   \n",
       "rs1000002      -0.5014     -0.2013     -0.3576      0.0717     -0.1686  ...   \n",
       "rs10000023      0.1033     -0.0169      0.0197      0.2101      0.0587  ...   \n",
       "rs1000003      -0.2013     -0.1730      0.1010      0.2544     -0.1112  ...   \n",
       "rs10000030     -0.3540      0.4442     -0.5963      0.0009     -0.4139  ...   \n",
       "...                ...         ...         ...         ...         ...  ...   \n",
       "VGXS34742       0.1807      0.2403      0.1901      0.3416      0.4133  ...   \n",
       "VGXS34743       0.1770      0.2181      0.0668      0.3055      0.4727  ...   \n",
       "VGXS34744      -0.0196     -0.0525     -0.0919      0.2553      0.2095  ...   \n",
       "VGXS34761       0.0827     -0.0679      0.1435      0.3957      0.2965  ...   \n",
       "VGXS35706      -0.2476      0.1102     -0.2108      0.3091      0.3059  ...   \n",
       "\n",
       "            GSM1196418  GSM1196419  GSM1196420  GSM1196421  GSM1196422  \\\n",
       "ID                                                                       \n",
       "rs1000000       0.0428     -0.0564      0.0930      0.1396      0.0075   \n",
       "rs1000002      -0.1097      0.1616     -0.2828     -0.2522      0.2335   \n",
       "rs10000023     -0.1189      0.0611      0.0374     -0.0478      0.1226   \n",
       "rs1000003       0.0022      0.0930     -0.1605     -0.1667      0.0316   \n",
       "rs10000030      0.3487      0.7111      0.5774      0.4109      0.3663   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "VGXS34742       0.2485      0.0686      0.1226     -0.0354     -0.1239   \n",
       "VGXS34743       0.2333      0.2030      0.1948      0.0367      0.0499   \n",
       "VGXS34744       0.0968     -0.0961     -0.1108     -0.4564     -0.4473   \n",
       "VGXS34761       0.2568      0.0615      0.0917     -0.1877     -0.1994   \n",
       "VGXS35706      -0.5293     -0.5513     -0.4255     -0.5476     -1.1190   \n",
       "\n",
       "            GSM1196423  GSM1196424  GSM1196425  GSM1196426  GSM1196427  \n",
       "ID                                                                      \n",
       "rs1000000       0.0441      0.0616      0.1482      0.0390      0.0873  \n",
       "rs1000002      -0.1791      0.2011      0.2515     -0.5280     -0.1804  \n",
       "rs10000023     -0.2908      0.1440      0.0704     -0.0951     -0.1550  \n",
       "rs1000003       0.0493     -0.1910      0.0574     -0.1625     -0.1756  \n",
       "rs10000030      0.2121      0.6078      0.8196      0.2831      0.8315  \n",
       "...                ...         ...         ...         ...         ...  \n",
       "VGXS34742      -0.3011      0.0032     -0.0218     -0.4481      0.0698  \n",
       "VGXS34743      -0.1797      0.1574     -0.0151     -0.3437      0.1108  \n",
       "VGXS34744      -0.7310     -0.1037     -0.5226     -0.7002     -0.2425  \n",
       "VGXS34761      -0.4965     -0.0968     -0.2921     -0.5193     -0.0352  \n",
       "VGXS35706      -1.4563     -0.9340     -0.8940     -1.8872     -0.3105  \n",
       "\n",
       "[710272 rows x 38 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[23], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m requires_gene_mapping \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_gene_mapping:\n\u001b[1;32m----> 4\u001b[0m     gene_annotation \u001b[38;5;241m=\u001b[39m \u001b[43mget_gene_annotation\u001b[49m\u001b[43m(\u001b[49m\u001b[43msoft_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      5\u001b[0m     gene_annotation_summary \u001b[38;5;241m=\u001b[39m preview_df(gene_annotation)\n\u001b[0;32m      6\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene_annotation_summary)\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:128\u001b[0m, in \u001b[0;36mget_gene_annotation\u001b[1;34m(file_path, prefixes)\u001b[0m\n\u001b[0;32m    126\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_gene_annotation\u001b[39m(file_path, prefixes\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m^\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    127\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Extract from a SOFT file the gene annotation data\"\"\"\u001b[39;00m\n\u001b[1;32m--> 128\u001b[0m     gene_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_content_by_prefix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprefixes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munselect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfile\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    129\u001b[0m \u001b[43m                                             \u001b[49m\u001b[43mreturn_df_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    130\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m gene_metadata[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:106\u001b[0m, in \u001b[0;36mfilter_content_by_prefix\u001b[1;34m(source, prefixes_a, prefixes_b, unselect, source_type, return_df_a, return_df_b)\u001b[0m\n\u001b[0;32m    104\u001b[0m filtered_content_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(filtered_lines_a)\n\u001b[0;32m    105\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m return_df_a:\n\u001b[1;32m--> 106\u001b[0m     filtered_content_a \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfiltered_content_a\u001b[49m\u001b[43m)\u001b[49m, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, low_memory\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, on_bad_lines\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mskip\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m    107\u001b[0m filtered_content_b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    108\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filtered_lines_b:\n",
      "\u001b[1;31mMemoryError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression analysis of metastatic adrenocortical tumors\"\n",
      "!Series_summary\t\"Background: Adrenocortical carcinoma (ACC) is a rare, often-aggressive neoplasm of the adrenal cortex, with a 14.5-month median overall survival. We asked whether tumors from patients with advanced or metastatic ACC would offer clues as to putative genes that might have critical roles in disease progression or in more aggressive disease biology.   Methods: We conducted comprehensive genomic and expression analyses of 43 ACCs.  Results: Copy number gains and losses matched that previously reported. We identified a median mutation rate of 3.38 per megabase (Mb), somewhat higher than in a previous study possibly related to the more advanced disease. The mutational signature was characterized by a predominance of C>T, C>A and T>C transitions. As in previously reports, only cancer genes TP53 (26%) and beta-catenin (CTNNB1, 14%) were mutated in more than 10% of samples. The TCGA-identified putative cancer genes MEN1 and PRKAR1A were found in low frequency â€“ 4.7% and 2.3%, respectively. Most of the mutations were in genes not implicated in the etiology or maintenance of cancer. Specifically, amongst the 38 genes that were mutated in more than 9% of samples, only four were represented in Tier 1 of the 576 COSMIC Cancer Gene Census (CCGC). Thus, 82% of genes found to have mutations likely have no role in the etiology or biology of ACC; while the role of the other 18%, if any, remains to be proven. Finally, the transcript length for the 38 most frequently mutated genes in ACC is statistically longer than the average of all coding genes, raising the question of whether transcript length in part determined mutation probability.   Conclusions: We conclude that the mutational and expression profiles of advanced and metastatic tumors is very similar to those from newly diagnosed patients â€“with very little in the way of genomic aberration to explain it. Our data and that in the previous analyses finds the rate of mutations in ACCs lower than that in other cancers and suggests an epigenetic basis for the disease should be the focus of future studies.\"\n",
      "!Series_summary\t\"The Affymetrix PrimeView platform was used for the gene expression profiling.\"\n",
      "!Series_overall_design\t\"Tumor samples embedded in OCT were sectioned, stained with hematoxylin and eosin and reviewed by a pathologist. RNA was extracted from 50-100 mg of tumor using the Qiagen miRNeasy Kit and then used for cDNA array analyses.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM4258059</th>\n",
       "      <th>GSM4258060</th>\n",
       "      <th>GSM4258061</th>\n",
       "      <th>GSM4258062</th>\n",
       "      <th>GSM4258063</th>\n",
       "      <th>GSM4258064</th>\n",
       "      <th>GSM4258065</th>\n",
       "      <th>GSM4258066</th>\n",
       "      <th>GSM4258067</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4258112</th>\n",
       "      <th>GSM4258113</th>\n",
       "      <th>GSM4258114</th>\n",
       "      <th>GSM4258115</th>\n",
       "      <th>GSM4258116</th>\n",
       "      <th>GSM4258117</th>\n",
       "      <th>GSM4258118</th>\n",
       "      <th>GSM4258119</th>\n",
       "      <th>GSM4258120</th>\n",
       "      <th>GSM4258121</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession GSM4258059 GSM4258060 GSM4258061 GSM4258062  \\\n",
       "0  !Sample_characteristics_ch1  gender: M  gender: F  gender: F  gender: F   \n",
       "\n",
       "  GSM4258063 GSM4258064 GSM4258065 GSM4258066 GSM4258067  ... GSM4258112  \\\n",
       "0  gender: F  gender: F  gender: F  gender: F  gender: M  ...  gender: F   \n",
       "\n",
       "  GSM4258113 GSM4258114 GSM4258115 GSM4258116 GSM4258117 GSM4258118  \\\n",
       "0  gender: M  gender: F  gender: M  gender: F  gender: M  gender: M   \n",
       "\n",
       "  GSM4258119 GSM4258120 GSM4258121  \n",
       "0  gender: F  gender: F  gender: M  \n",
       "\n",
       "[1 rows x 64 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only gender --> SKIP\n",
    "cohort = accession_num = \"GSE143383\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Expression data from human metastatic adrenocortical carcinoma\"\n",
      "!Series_summary\t\"CXCR4 expression by metastatic adrenocortical carcinoma is heterogeneous among patients and among lesions\"\n",
      "!Series_summary\t\"We used microarrays for 57 ACC metastases from 42 patients to evaluate gene expression in different lesions from same patients and over time, focusing on CXCR4 expression and other genes correlating with CXCR4 expression\"\n",
      "!Series_overall_design\t\"57 ACC metastases from 42 patients were used for RNA extraction and hybridization on Affymetrix microarrays. We sought to obtain data on CXCR4 expression by ACC metastases. Multiple lesion samples were aquired for 9 of the patients, labeled a thru i. Single samples were aquired from the other subjects.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2411058</th>\n",
       "      <th>GSM2411059</th>\n",
       "      <th>GSM2411060</th>\n",
       "      <th>GSM2411061</th>\n",
       "      <th>GSM2411062</th>\n",
       "      <th>GSM2411063</th>\n",
       "      <th>GSM2411064</th>\n",
       "      <th>GSM2411065</th>\n",
       "      <th>GSM2411066</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2411111</th>\n",
       "      <th>GSM2411112</th>\n",
       "      <th>GSM2411113</th>\n",
       "      <th>GSM2411114</th>\n",
       "      <th>GSM2411115</th>\n",
       "      <th>GSM2411116</th>\n",
       "      <th>GSM2411117</th>\n",
       "      <th>GSM2411118</th>\n",
       "      <th>GSM2411119</th>\n",
       "      <th>GSM2411120</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "      <td>tissue: adrenocortical carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>study: 426</td>\n",
       "      <td>...</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "      <td>study: 920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "      <td>condition: tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>acc_num: 1</td>\n",
       "      <td>acc_num: 6</td>\n",
       "      <td>acc_num: 7</td>\n",
       "      <td>acc_num: 8</td>\n",
       "      <td>acc_num: 9</td>\n",
       "      <td>acc_num: 11</td>\n",
       "      <td>acc_num: 13</td>\n",
       "      <td>acc_num: 14</td>\n",
       "      <td>acc_num: 15</td>\n",
       "      <td>...</td>\n",
       "      <td>acc_num: 28b</td>\n",
       "      <td>acc_num: 29b</td>\n",
       "      <td>acc_num: 31b</td>\n",
       "      <td>acc_num: 30newb</td>\n",
       "      <td>acc_num: 32newb</td>\n",
       "      <td>acc_num: 3b</td>\n",
       "      <td>acc_num: 4b</td>\n",
       "      <td>acc_num: 5b</td>\n",
       "      <td>acc_num: 7b</td>\n",
       "      <td>acc_num: 8b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>patient: a</td>\n",
       "      <td>patient: b</td>\n",
       "      <td>patient: c</td>\n",
       "      <td>patient: b</td>\n",
       "      <td>patient: d</td>\n",
       "      <td>patient: A_16</td>\n",
       "      <td>patient: A_17</td>\n",
       "      <td>patient: A_20</td>\n",
       "      <td>patient: A_22</td>\n",
       "      <td>...</td>\n",
       "      <td>patient: i</td>\n",
       "      <td>patient: i</td>\n",
       "      <td>patient: i</td>\n",
       "      <td>patient: i</td>\n",
       "      <td>patient: i</td>\n",
       "      <td>patient: B_32_1</td>\n",
       "      <td>patient: B_4</td>\n",
       "      <td>patient: B_5</td>\n",
       "      <td>patient: B_7</td>\n",
       "      <td>patient: B_8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                        GSM2411058  \\\n",
       "0  !Sample_characteristics_ch1  tissue: adrenocortical carcinoma   \n",
       "1  !Sample_characteristics_ch1                        study: 426   \n",
       "2  !Sample_characteristics_ch1                  condition: tumor   \n",
       "3  !Sample_characteristics_ch1                        acc_num: 1   \n",
       "4  !Sample_characteristics_ch1                        patient: a   \n",
       "\n",
       "                         GSM2411059                        GSM2411060  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 426                        study: 426   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                        acc_num: 6                        acc_num: 7   \n",
       "4                        patient: b                        patient: c   \n",
       "\n",
       "                         GSM2411061                        GSM2411062  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 426                        study: 426   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                        acc_num: 8                        acc_num: 9   \n",
       "4                        patient: b                        patient: d   \n",
       "\n",
       "                         GSM2411063                        GSM2411064  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 426                        study: 426   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                       acc_num: 11                       acc_num: 13   \n",
       "4                     patient: A_16                     patient: A_17   \n",
       "\n",
       "                         GSM2411065                        GSM2411066  ...  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma  ...   \n",
       "1                        study: 426                        study: 426  ...   \n",
       "2                  condition: tumor                  condition: tumor  ...   \n",
       "3                       acc_num: 14                       acc_num: 15  ...   \n",
       "4                     patient: A_20                     patient: A_22  ...   \n",
       "\n",
       "                         GSM2411111                        GSM2411112  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 920                        study: 920   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                      acc_num: 28b                      acc_num: 29b   \n",
       "4                        patient: i                        patient: i   \n",
       "\n",
       "                         GSM2411113                        GSM2411114  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 920                        study: 920   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                      acc_num: 31b                   acc_num: 30newb   \n",
       "4                        patient: i                        patient: i   \n",
       "\n",
       "                         GSM2411115                        GSM2411116  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 920                        study: 920   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                   acc_num: 32newb                       acc_num: 3b   \n",
       "4                        patient: i                   patient: B_32_1   \n",
       "\n",
       "                         GSM2411117                        GSM2411118  \\\n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma   \n",
       "1                        study: 920                        study: 920   \n",
       "2                  condition: tumor                  condition: tumor   \n",
       "3                       acc_num: 4b                       acc_num: 5b   \n",
       "4                      patient: B_4                      patient: B_5   \n",
       "\n",
       "                         GSM2411119                        GSM2411120  \n",
       "0  tissue: adrenocortical carcinoma  tissue: adrenocortical carcinoma  \n",
       "1                        study: 920                        study: 920  \n",
       "2                  condition: tumor                  condition: tumor  \n",
       "3                       acc_num: 7b                       acc_num: 8b  \n",
       "4                      patient: B_7                      patient: B_8  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished????\n",
    "\n",
    "cohort = accession_num = \"GSE90713\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'condition: tumor',\n",
       "       'condition: normal'], dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2411058</th>\n",
       "      <th>GSM2411059</th>\n",
       "      <th>GSM2411060</th>\n",
       "      <th>GSM2411061</th>\n",
       "      <th>GSM2411062</th>\n",
       "      <th>GSM2411063</th>\n",
       "      <th>GSM2411064</th>\n",
       "      <th>GSM2411065</th>\n",
       "      <th>GSM2411066</th>\n",
       "      <th>GSM2411067</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2411111</th>\n",
       "      <th>GSM2411112</th>\n",
       "      <th>GSM2411113</th>\n",
       "      <th>GSM2411114</th>\n",
       "      <th>GSM2411115</th>\n",
       "      <th>GSM2411116</th>\n",
       "      <th>GSM2411117</th>\n",
       "      <th>GSM2411118</th>\n",
       "      <th>GSM2411119</th>\n",
       "      <th>GSM2411120</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11715100_at</th>\n",
       "      <td>4.841518</td>\n",
       "      <td>4.947266</td>\n",
       "      <td>4.903320</td>\n",
       "      <td>5.007812</td>\n",
       "      <td>5.004883</td>\n",
       "      <td>5.092773</td>\n",
       "      <td>4.888672</td>\n",
       "      <td>4.941406</td>\n",
       "      <td>5.827148</td>\n",
       "      <td>4.951172</td>\n",
       "      <td>...</td>\n",
       "      <td>4.987305</td>\n",
       "      <td>4.831055</td>\n",
       "      <td>4.747070</td>\n",
       "      <td>5.110352</td>\n",
       "      <td>4.948242</td>\n",
       "      <td>4.866211</td>\n",
       "      <td>5.023438</td>\n",
       "      <td>5.068359</td>\n",
       "      <td>5.014648</td>\n",
       "      <td>4.917969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11715101_s_at</th>\n",
       "      <td>5.293186</td>\n",
       "      <td>5.039062</td>\n",
       "      <td>5.093750</td>\n",
       "      <td>5.549805</td>\n",
       "      <td>5.667969</td>\n",
       "      <td>5.346680</td>\n",
       "      <td>5.472656</td>\n",
       "      <td>5.256836</td>\n",
       "      <td>5.132812</td>\n",
       "      <td>5.692383</td>\n",
       "      <td>...</td>\n",
       "      <td>5.365234</td>\n",
       "      <td>5.356445</td>\n",
       "      <td>5.077148</td>\n",
       "      <td>5.928711</td>\n",
       "      <td>5.514648</td>\n",
       "      <td>5.224609</td>\n",
       "      <td>5.111328</td>\n",
       "      <td>5.124023</td>\n",
       "      <td>5.298828</td>\n",
       "      <td>5.390625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11715102_x_at</th>\n",
       "      <td>4.914528</td>\n",
       "      <td>5.022461</td>\n",
       "      <td>4.991211</td>\n",
       "      <td>5.166992</td>\n",
       "      <td>5.165039</td>\n",
       "      <td>5.042969</td>\n",
       "      <td>4.958008</td>\n",
       "      <td>5.041992</td>\n",
       "      <td>5.163086</td>\n",
       "      <td>5.054688</td>\n",
       "      <td>...</td>\n",
       "      <td>5.144531</td>\n",
       "      <td>4.923828</td>\n",
       "      <td>4.839844</td>\n",
       "      <td>5.366211</td>\n",
       "      <td>5.056641</td>\n",
       "      <td>4.867188</td>\n",
       "      <td>5.034180</td>\n",
       "      <td>5.160156</td>\n",
       "      <td>5.103516</td>\n",
       "      <td>5.023438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11715103_x_at</th>\n",
       "      <td>5.504681</td>\n",
       "      <td>5.315430</td>\n",
       "      <td>5.303711</td>\n",
       "      <td>5.375000</td>\n",
       "      <td>5.358398</td>\n",
       "      <td>5.333984</td>\n",
       "      <td>5.676758</td>\n",
       "      <td>5.392578</td>\n",
       "      <td>6.496094</td>\n",
       "      <td>5.499023</td>\n",
       "      <td>...</td>\n",
       "      <td>5.378906</td>\n",
       "      <td>5.464844</td>\n",
       "      <td>5.498047</td>\n",
       "      <td>5.693359</td>\n",
       "      <td>5.537109</td>\n",
       "      <td>5.397461</td>\n",
       "      <td>5.589844</td>\n",
       "      <td>6.132812</td>\n",
       "      <td>5.736328</td>\n",
       "      <td>5.473633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11715104_s_at</th>\n",
       "      <td>5.056718</td>\n",
       "      <td>5.155273</td>\n",
       "      <td>5.315430</td>\n",
       "      <td>5.049805</td>\n",
       "      <td>5.089844</td>\n",
       "      <td>5.458008</td>\n",
       "      <td>5.041016</td>\n",
       "      <td>5.130859</td>\n",
       "      <td>5.394531</td>\n",
       "      <td>5.101562</td>\n",
       "      <td>...</td>\n",
       "      <td>5.104492</td>\n",
       "      <td>5.126953</td>\n",
       "      <td>5.010742</td>\n",
       "      <td>5.333008</td>\n",
       "      <td>5.175781</td>\n",
       "      <td>4.900391</td>\n",
       "      <td>5.298828</td>\n",
       "      <td>5.469727</td>\n",
       "      <td>5.416992</td>\n",
       "      <td>5.142578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>5.424495</td>\n",
       "      <td>6.557617</td>\n",
       "      <td>7.363281</td>\n",
       "      <td>6.172852</td>\n",
       "      <td>6.250977</td>\n",
       "      <td>5.279297</td>\n",
       "      <td>6.483398</td>\n",
       "      <td>5.523438</td>\n",
       "      <td>5.412109</td>\n",
       "      <td>5.427734</td>\n",
       "      <td>...</td>\n",
       "      <td>6.426758</td>\n",
       "      <td>6.051758</td>\n",
       "      <td>8.982422</td>\n",
       "      <td>5.268555</td>\n",
       "      <td>5.638672</td>\n",
       "      <td>6.011719</td>\n",
       "      <td>7.086914</td>\n",
       "      <td>6.926758</td>\n",
       "      <td>6.325195</td>\n",
       "      <td>5.761719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>6.360383</td>\n",
       "      <td>8.201172</td>\n",
       "      <td>9.500000</td>\n",
       "      <td>7.684570</td>\n",
       "      <td>7.594727</td>\n",
       "      <td>5.778320</td>\n",
       "      <td>8.029297</td>\n",
       "      <td>6.305664</td>\n",
       "      <td>6.199219</td>\n",
       "      <td>5.916992</td>\n",
       "      <td>...</td>\n",
       "      <td>7.393555</td>\n",
       "      <td>6.994141</td>\n",
       "      <td>10.042969</td>\n",
       "      <td>6.012695</td>\n",
       "      <td>6.388672</td>\n",
       "      <td>7.298828</td>\n",
       "      <td>8.451172</td>\n",
       "      <td>8.498047</td>\n",
       "      <td>7.401367</td>\n",
       "      <td>6.551758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>4.642501</td>\n",
       "      <td>4.609375</td>\n",
       "      <td>4.632812</td>\n",
       "      <td>4.607422</td>\n",
       "      <td>4.633789</td>\n",
       "      <td>4.631836</td>\n",
       "      <td>4.649414</td>\n",
       "      <td>4.630859</td>\n",
       "      <td>4.643555</td>\n",
       "      <td>4.661133</td>\n",
       "      <td>...</td>\n",
       "      <td>4.631836</td>\n",
       "      <td>4.645508</td>\n",
       "      <td>4.626953</td>\n",
       "      <td>4.680664</td>\n",
       "      <td>4.624023</td>\n",
       "      <td>4.647461</td>\n",
       "      <td>4.639648</td>\n",
       "      <td>4.665039</td>\n",
       "      <td>4.630859</td>\n",
       "      <td>4.632812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>4.758898</td>\n",
       "      <td>4.708008</td>\n",
       "      <td>4.736328</td>\n",
       "      <td>4.763672</td>\n",
       "      <td>4.761719</td>\n",
       "      <td>4.729492</td>\n",
       "      <td>4.740234</td>\n",
       "      <td>4.726562</td>\n",
       "      <td>4.725586</td>\n",
       "      <td>4.826172</td>\n",
       "      <td>...</td>\n",
       "      <td>4.691406</td>\n",
       "      <td>4.776367</td>\n",
       "      <td>4.668945</td>\n",
       "      <td>4.723633</td>\n",
       "      <td>4.730469</td>\n",
       "      <td>4.751953</td>\n",
       "      <td>4.738281</td>\n",
       "      <td>4.726562</td>\n",
       "      <td>4.751953</td>\n",
       "      <td>4.698242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>4.699637</td>\n",
       "      <td>4.697266</td>\n",
       "      <td>4.683594</td>\n",
       "      <td>4.685547</td>\n",
       "      <td>4.723633</td>\n",
       "      <td>4.711914</td>\n",
       "      <td>4.742188</td>\n",
       "      <td>4.699219</td>\n",
       "      <td>4.817383</td>\n",
       "      <td>4.736328</td>\n",
       "      <td>...</td>\n",
       "      <td>4.691406</td>\n",
       "      <td>4.744141</td>\n",
       "      <td>4.679688</td>\n",
       "      <td>4.688477</td>\n",
       "      <td>4.698242</td>\n",
       "      <td>4.731445</td>\n",
       "      <td>4.707031</td>\n",
       "      <td>4.702148</td>\n",
       "      <td>4.674805</td>\n",
       "      <td>4.691406</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49395 rows Ã— 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM2411058  GSM2411059  GSM2411060  GSM2411061  GSM2411062  \\\n",
       "ID                                                                            \n",
       "11715100_at        4.841518    4.947266    4.903320    5.007812    5.004883   \n",
       "11715101_s_at      5.293186    5.039062    5.093750    5.549805    5.667969   \n",
       "11715102_x_at      4.914528    5.022461    4.991211    5.166992    5.165039   \n",
       "11715103_x_at      5.504681    5.315430    5.303711    5.375000    5.358398   \n",
       "11715104_s_at      5.056718    5.155273    5.315430    5.049805    5.089844   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     5.424495    6.557617    7.363281    6.172852    6.250977   \n",
       "AFFX-ThrX-M_at     6.360383    8.201172    9.500000    7.684570    7.594727   \n",
       "AFFX-TrpnX-3_at    4.642501    4.609375    4.632812    4.607422    4.633789   \n",
       "AFFX-TrpnX-5_at    4.758898    4.708008    4.736328    4.763672    4.761719   \n",
       "AFFX-TrpnX-M_at    4.699637    4.697266    4.683594    4.685547    4.723633   \n",
       "\n",
       "                 GSM2411063  GSM2411064  GSM2411065  GSM2411066  GSM2411067  \\\n",
       "ID                                                                            \n",
       "11715100_at        5.092773    4.888672    4.941406    5.827148    4.951172   \n",
       "11715101_s_at      5.346680    5.472656    5.256836    5.132812    5.692383   \n",
       "11715102_x_at      5.042969    4.958008    5.041992    5.163086    5.054688   \n",
       "11715103_x_at      5.333984    5.676758    5.392578    6.496094    5.499023   \n",
       "11715104_s_at      5.458008    5.041016    5.130859    5.394531    5.101562   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     5.279297    6.483398    5.523438    5.412109    5.427734   \n",
       "AFFX-ThrX-M_at     5.778320    8.029297    6.305664    6.199219    5.916992   \n",
       "AFFX-TrpnX-3_at    4.631836    4.649414    4.630859    4.643555    4.661133   \n",
       "AFFX-TrpnX-5_at    4.729492    4.740234    4.726562    4.725586    4.826172   \n",
       "AFFX-TrpnX-M_at    4.711914    4.742188    4.699219    4.817383    4.736328   \n",
       "\n",
       "                 ...  GSM2411111  GSM2411112  GSM2411113  GSM2411114  \\\n",
       "ID               ...                                                   \n",
       "11715100_at      ...    4.987305    4.831055    4.747070    5.110352   \n",
       "11715101_s_at    ...    5.365234    5.356445    5.077148    5.928711   \n",
       "11715102_x_at    ...    5.144531    4.923828    4.839844    5.366211   \n",
       "11715103_x_at    ...    5.378906    5.464844    5.498047    5.693359   \n",
       "11715104_s_at    ...    5.104492    5.126953    5.010742    5.333008   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at   ...    6.426758    6.051758    8.982422    5.268555   \n",
       "AFFX-ThrX-M_at   ...    7.393555    6.994141   10.042969    6.012695   \n",
       "AFFX-TrpnX-3_at  ...    4.631836    4.645508    4.626953    4.680664   \n",
       "AFFX-TrpnX-5_at  ...    4.691406    4.776367    4.668945    4.723633   \n",
       "AFFX-TrpnX-M_at  ...    4.691406    4.744141    4.679688    4.688477   \n",
       "\n",
       "                 GSM2411115  GSM2411116  GSM2411117  GSM2411118  GSM2411119  \\\n",
       "ID                                                                            \n",
       "11715100_at        4.948242    4.866211    5.023438    5.068359    5.014648   \n",
       "11715101_s_at      5.514648    5.224609    5.111328    5.124023    5.298828   \n",
       "11715102_x_at      5.056641    4.867188    5.034180    5.160156    5.103516   \n",
       "11715103_x_at      5.537109    5.397461    5.589844    6.132812    5.736328   \n",
       "11715104_s_at      5.175781    4.900391    5.298828    5.469727    5.416992   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     5.638672    6.011719    7.086914    6.926758    6.325195   \n",
       "AFFX-ThrX-M_at     6.388672    7.298828    8.451172    8.498047    7.401367   \n",
       "AFFX-TrpnX-3_at    4.624023    4.647461    4.639648    4.665039    4.630859   \n",
       "AFFX-TrpnX-5_at    4.730469    4.751953    4.738281    4.726562    4.751953   \n",
       "AFFX-TrpnX-M_at    4.698242    4.731445    4.707031    4.702148    4.674805   \n",
       "\n",
       "                 GSM2411120  \n",
       "ID                           \n",
       "11715100_at        4.917969  \n",
       "11715101_s_at      5.390625  \n",
       "11715102_x_at      5.023438  \n",
       "11715103_x_at      5.473633  \n",
       "11715104_s_at      5.142578  \n",
       "...                     ...  \n",
       "AFFX-ThrX-5_at     5.761719  \n",
       "AFFX-ThrX-M_at     6.551758  \n",
       "AFFX-TrpnX-3_at    4.632812  \n",
       "AFFX-TrpnX-5_at    4.698242  \n",
       "AFFX-TrpnX-M_at    4.691406  \n",
       "\n",
       "[49395 rows x 63 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['11715100_at', '11715101_s_at', '11715102_x_at', '11715103_x_at', '11715104_s_at'], 'GeneChip Array': ['Human Genome PrimeView Array', 'Human Genome PrimeView Array', 'Human Genome PrimeView Array', 'Human Genome PrimeView Array', 'Human Genome PrimeView Array'], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['30-Mar-16', '30-Mar-16', '30-Mar-16', '30-Mar-16', '30-Mar-16'], 'Sequence Type': ['Consensus sequence', 'Consensus sequence', 'Consensus sequence', 'Consensus sequence', 'Consensus sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'Affymetrix Proprietary Database', 'Affymetrix Proprietary Database', 'Affymetrix Proprietary Database', 'Affymetrix Proprietary Database'], 'Transcript ID(Array Design)': ['g21264570', 'g21264570', 'g21264570', 'g22748780', 'g30039713'], 'Target Description': ['g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g21264570 /REP_ORG=Homo sapiens', 'g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g21264570 /REP_ORG=Homo sapiens', 'g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g21264570 /REP_ORG=Homo sapiens', 'g22748780 /TID=g22748780 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g22748780 /REP_ORG=Homo sapiens', 'g30039713 /TID=g30039713 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g30039713 /REP_ORG=Homo sapiens'], 'GB_ACC': [nan, nan, nan, nan, nan], 'GI': [21264570.0, 21264570.0, 21264570.0, 22748780.0, 30039713.0], 'Representative Public ID': ['g21264570', 'g21264570', 'g21264570', 'g22748780', 'g30039713'], 'Archival UniGene Cluster': ['---', '---', '---', '---', '---'], 'UniGene ID': ['Hs.247813', 'Hs.247813', 'Hs.247813', 'Hs.465643', 'Hs.352515'], 'Genome Version': ['February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)'], 'Alignments': ['chr6:26271145-26271612 (-) // 100.0 // p22.2', 'chr6:26271145-26271612 (-) // 100.0 // p22.2', 'chr6:26271145-26271612 (-) // 100.0 // p22.2', 'chr19:4639529-5145579 (+) // 48.53 // p13.3', 'chr17:72920369-72929640 (+) // 100.0 // q25.1'], 'Gene Title': ['histone cluster 1, H3g', 'histone cluster 1, H3g', 'histone cluster 1, H3g', 'tumor necrosis factor, alpha-induced protein 8-like 1', 'otopetrin 2'], 'Gene Symbol': ['HIST1H3G', 'HIST1H3G', 'HIST1H3G', 'TNFAIP8L1', 'OTOP2'], 'Chromosomal Location': ['chr6p22.2', 'chr6p22.2', 'chr6p22.2', 'chr19p13.3', 'chr17q25.1'], 'Unigene Cluster Type': ['full length', 'full length', 'full length', 'full length', 'full length'], 'Ensembl': ['ENSG00000273983 /// OTTHUMG00000014436', 'ENSG00000273983 /// OTTHUMG00000014436', 'ENSG00000273983 /// OTTHUMG00000014436', 'ENSG00000185361 /// OTTHUMG00000182013', 'ENSG00000183034 /// OTTHUMG00000179215'], 'Entrez Gene': ['8355', '8355', '8355', '126282', '92736'], 'SwissProt': ['P68431', 'P68431', 'P68431', 'Q8WVP5', 'Q7RTS6'], 'EC': ['---', '---', '---', '---', '---'], 'OMIM': ['602815', '602815', '602815', '615869', '607827'], 'RefSeq Protein ID': ['NP_003525', 'NP_003525', 'NP_003525', 'NP_001161414 /// NP_689575 /// XP_005259544 /// XP_011525982', 'NP_835454 /// XP_011523781'], 'RefSeq Transcript ID': ['NM_003534', 'NM_003534', 'NM_003534', 'NM_001167942 /// NM_152362 /// XM_005259487 /// XM_011527680', 'NM_178160 /// XM_011525479'], 'Gene Ontology Biological Process': ['0000183 // chromatin silencing at rDNA // traceable author statement /// 0002230 // positive regulation of defense response to virus by host // inferred from mutant phenotype /// 0006325 // chromatin organization // traceable author statement /// 0006334 // nucleosome assembly // inferred from direct assay /// 0006334 // nucleosome assembly // inferred from mutant phenotype /// 0006335 // DNA replication-dependent nucleosome assembly // inferred from direct assay /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0010467 // gene expression // traceable author statement /// 0031047 // gene silencing by RNA // traceable author statement /// 0032776 // DNA methylation on cytosine // traceable author statement /// 0040029 // regulation of gene expression, epigenetic // traceable author statement /// 0044267 // cellular protein metabolic process // traceable author statement /// 0045814 // negative regulation of gene expression, epigenetic // traceable author statement /// 0051290 // protein heterotetramerization // inferred from direct assay /// 0060968 // regulation of gene silencing // inferred from direct assay /// 0098792 // xenophagy // inferred from mutant phenotype', '0000183 // chromatin silencing at rDNA // traceable author statement /// 0002230 // positive regulation of defense response to virus by host // inferred from mutant phenotype /// 0006325 // chromatin organization // traceable author statement /// 0006334 // nucleosome assembly // inferred from direct assay /// 0006334 // nucleosome assembly // inferred from mutant phenotype /// 0006335 // DNA replication-dependent nucleosome assembly // inferred from direct assay /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0010467 // gene expression // traceable author statement /// 0031047 // gene silencing by RNA // traceable author statement /// 0032776 // DNA methylation on cytosine // traceable author statement /// 0040029 // regulation of gene expression, epigenetic // traceable author statement /// 0044267 // cellular protein metabolic process // traceable author statement /// 0045814 // negative regulation of gene expression, epigenetic // traceable author statement /// 0051290 // protein heterotetramerization // inferred from direct assay /// 0060968 // regulation of gene silencing // inferred from direct assay /// 0098792 // xenophagy // inferred from mutant phenotype', '0000183 // chromatin silencing at rDNA // traceable author statement /// 0002230 // positive regulation of defense response to virus by host // inferred from mutant phenotype /// 0006325 // chromatin organization // traceable author statement /// 0006334 // nucleosome assembly // inferred from direct assay /// 0006334 // nucleosome assembly // inferred from mutant phenotype /// 0006335 // DNA replication-dependent nucleosome assembly // inferred from direct assay /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0010467 // gene expression // traceable author statement /// 0031047 // gene silencing by RNA // traceable author statement /// 0032776 // DNA methylation on cytosine // traceable author statement /// 0040029 // regulation of gene expression, epigenetic // traceable author statement /// 0044267 // cellular protein metabolic process // traceable author statement /// 0045814 // negative regulation of gene expression, epigenetic // traceable author statement /// 0051290 // protein heterotetramerization // inferred from direct assay /// 0060968 // regulation of gene silencing // inferred from direct assay /// 0098792 // xenophagy // inferred from mutant phenotype', '0032007 // negative regulation of TOR signaling // not recorded /// 0032007 // negative regulation of TOR signaling // inferred from sequence or structural similarity', '---'], 'Gene Ontology Cellular Component': ['0000228 // nuclear chromosome // inferred from direct assay /// 0000786 // nucleosome // inferred from direct assay /// 0000788 // nuclear nucleosome // inferred from direct assay /// 0005576 // extracellular region // traceable author statement /// 0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // traceable author statement /// 0005694 // chromosome // inferred from electronic annotation /// 0016020 // membrane // inferred from direct assay /// 0043234 // protein complex // inferred from direct assay /// 0070062 // extracellular exosome // inferred from direct assay', '0000228 // nuclear chromosome // inferred from direct assay /// 0000786 // nucleosome // inferred from direct assay /// 0000788 // nuclear nucleosome // inferred from direct assay /// 0005576 // extracellular region // traceable author statement /// 0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // traceable author statement /// 0005694 // chromosome // inferred from electronic annotation /// 0016020 // membrane // inferred from direct assay /// 0043234 // protein complex // inferred from direct assay /// 0070062 // extracellular exosome // inferred from direct assay', '0000228 // nuclear chromosome // inferred from direct assay /// 0000786 // nucleosome // inferred from direct assay /// 0000788 // nuclear nucleosome // inferred from direct assay /// 0005576 // extracellular region // traceable author statement /// 0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // traceable author statement /// 0005694 // chromosome // inferred from electronic annotation /// 0016020 // membrane // inferred from direct assay /// 0043234 // protein complex // inferred from direct assay /// 0070062 // extracellular exosome // inferred from direct assay', '0005737 // cytoplasm // not recorded /// 0005737 // cytoplasm // inferred from sequence or structural similarity', '0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation'], 'Gene Ontology Molecular Function': ['0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0042393 // histone binding // inferred from physical interaction /// 0046982 // protein heterodimerization activity // inferred from electronic annotation', '0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0042393 // histone binding // inferred from physical interaction /// 0046982 // protein heterodimerization activity // inferred from electronic annotation', '0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0042393 // histone binding // inferred from physical interaction /// 0046982 // protein heterodimerization activity // inferred from electronic annotation', '0005515 // protein binding // inferred from physical interaction', '---'], 'Pathway': ['---', '---', '---', '---', '---'], 'InterPro': ['IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 /// IPR007125 // Histone H2A/H2B/H3 // 1.7E-37', 'IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 /// IPR007125 // Histone H2A/H2B/H3 // 1.7E-37', 'IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 /// IPR007125 // Histone H2A/H2B/H3 // 1.7E-37', 'IPR008477 // Protein of unknown function DUF758 // 8.4E-86 /// IPR008477 // Protein of unknown function DUF758 // 6.8E-90', 'IPR004878 // Otopetrin // 9.4E-43 /// IPR004878 // Otopetrin // 9.4E-43 /// IPR004878 // Otopetrin // 9.4E-43 /// IPR004878 // Otopetrin // 3.9E-18 /// IPR004878 // Otopetrin // 3.8E-20 /// IPR004878 // Otopetrin // 5.2E-16'], 'Annotation Description': ['This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 4 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 4 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 4 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 9 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 6 transcripts. // false // Matching Probes // A'], 'Annotation Transcript Cluster': ['ENST00000614378(11),NM_003534(11),OTTHUMT00000040099(11),uc003nhi.3', 'ENST00000614378(11),NM_003534(11),OTTHUMT00000040099(11),uc003nhi.3', 'ENST00000614378(11),NM_003534(11),OTTHUMT00000040099(11),uc003nhi.3', 'BC017672(11),BC044250(9),ENST00000327473(11),ENST00000536716(11),NM_001167942(11),NM_152362(11),OTTHUMT00000458662(11),uc002max.3,uc021une.1', 'ENST00000331427(11),ENST00000580223(11),NM_178160(11),OTTHUMT00000445306(11),uc010wrp.2,XM_011525479(11)'], 'Transcript Assignments': ['ENST00000614378 // ensembl_havana_transcript:known chromosome:GRCh38:6:26269405:26271815:-1 gene:ENSG00000273983 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_003534 // Homo sapiens histone cluster 1, H3g (HIST1H3G), mRNA. // refseq // 11 // --- /// OTTHUMT00000040099 // otter:known chromosome:VEGA61:6:26269405:26271815:-1 gene:OTTHUMG00000014436 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc003nhi.3 // --- // ucsc_genes // 11 // ---', 'ENST00000614378 // ensembl_havana_transcript:known chromosome:GRCh38:6:26269405:26271815:-1 gene:ENSG00000273983 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// GENSCAN00000029819 // cdna:genscan chromosome:GRCh38:6:26270974:26271384:-1 transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_003534 // Homo sapiens histone cluster 1, H3g (HIST1H3G), mRNA. // refseq // 11 // --- /// OTTHUMT00000040099 // otter:known chromosome:VEGA61:6:26269405:26271815:-1 gene:OTTHUMG00000014436 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc003nhi.3 // --- // ucsc_genes // 11 // ---', 'ENST00000614378 // ensembl_havana_transcript:known chromosome:GRCh38:6:26269405:26271815:-1 gene:ENSG00000273983 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_003534 // Homo sapiens histone cluster 1, H3g (HIST1H3G), mRNA. // refseq // 11 // --- /// OTTHUMT00000040099 // otter:known chromosome:VEGA61:6:26269405:26271815:-1 gene:OTTHUMG00000014436 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc003nhi.3 // --- // ucsc_genes // 11 // ---', 'BC017672 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1, mRNA (cDNA clone MGC:17791 IMAGE:3885999), complete cds. // gb // 11 // --- /// BC044250 // accn=BC044250 class=mRNAlike lncRNA name=Human lncRNA ref=JounralRNA transcriptId=673 cpcScore=-0.1526100 cnci=-0.1238602 // noncode // 9 // --- /// BC044250 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1, mRNA (cDNA clone IMAGE:5784807). // gb // 9 // --- /// ENST00000327473 // ensembl_havana_transcript:known chromosome:GRCh38:19:4639518:4655568:1 gene:ENSG00000185361 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000536716 // ensembl:known chromosome:GRCh38:19:4640017:4655568:1 gene:ENSG00000185361 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_001167942 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1 (TNFAIP8L1), transcript variant 1, mRNA. // refseq // 11 // --- /// NM_152362 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1 (TNFAIP8L1), transcript variant 2, mRNA. // refseq // 11 // --- /// NONHSAT060631 // Non-coding transcript identified by NONCODE: Exonic // noncode // 9 // --- /// OTTHUMT00000458662 // otter:known chromosome:VEGA61:19:4639518:4655568:1 gene:OTTHUMG00000182013 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc002max.3 // --- // ucsc_genes // 11 // --- /// uc021une.1 // --- // ucsc_genes // 11 // ---', 'ENST00000331427 // ensembl:known chromosome:GRCh38:17:74924275:74933911:1 gene:ENSG00000183034 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000580223 // havana:known chromosome:GRCh38:17:74924603:74933912:1 gene:ENSG00000183034 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// GENSCAN00000013715 // cdna:genscan chromosome:GRCh38:17:74924633:74933545:1 transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_178160 // Homo sapiens otopetrin 2 (OTOP2), mRNA. // refseq // 11 // --- /// OTTHUMT00000445306 // otter:known chromosome:VEGA61:17:74924603:74933912:1 gene:OTTHUMG00000179215 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc010wrp.2 // --- // ucsc_genes // 11 // --- /// XM_011525479 // PREDICTED: Homo sapiens otopetrin 2 (OTOP2), transcript variant X1, mRNA. // refseq // 11 // ---'], 'Annotation Notes': ['---', '---', 'GENSCAN00000029819 // ensembl // 4 // Cross Hyb Matching Probes', '---', '---'], 'SPOT_ID': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['11715100_at', '11715101_s_at', '11715102_x_at', '11715103_x_at', '11715104_s_at', '11715105_at', '11715106_x_at', '11715107_s_at', '11715108_x_at', '11715109_at', '11715110_at', '11715111_s_at', '11715112_at', '11715113_x_at', '11715114_x_at', '11715115_s_at', '11715116_s_at', '11715117_x_at', '11715118_s_at', '11715119_s_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['11715100_at', '11715101_s_at', '11715102_x_at', '11715103_x_at', '11715104_s_at'], 'GeneChip Array': ['Human Genome PrimeView Array', 'Human Genome PrimeView Array', 'Human Genome PrimeView Array', 'Human Genome PrimeView Array', 'Human Genome PrimeView Array'], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['30-Mar-16', '30-Mar-16', '30-Mar-16', '30-Mar-16', '30-Mar-16'], 'Sequence Type': ['Consensus sequence', 'Consensus sequence', 'Consensus sequence', 'Consensus sequence', 'Consensus sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'Affymetrix Proprietary Database', 'Affymetrix Proprietary Database', 'Affymetrix Proprietary Database', 'Affymetrix Proprietary Database'], 'Transcript ID(Array Design)': ['g21264570', 'g21264570', 'g21264570', 'g22748780', 'g30039713'], 'Target Description': ['g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g21264570 /REP_ORG=Homo sapiens', 'g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g21264570 /REP_ORG=Homo sapiens', 'g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g21264570 /REP_ORG=Homo sapiens', 'g22748780 /TID=g22748780 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g22748780 /REP_ORG=Homo sapiens', 'g30039713 /TID=g30039713 /CNT=1 /FEA=FLmRNA /TIER=FL /STK=0 /DEF=g30039713 /REP_ORG=Homo sapiens'], 'GB_ACC': [nan, nan, nan, nan, nan], 'GI': [21264570.0, 21264570.0, 21264570.0, 22748780.0, 30039713.0], 'Representative Public ID': ['g21264570', 'g21264570', 'g21264570', 'g22748780', 'g30039713'], 'Archival UniGene Cluster': ['---', '---', '---', '---', '---'], 'UniGene ID': ['Hs.247813', 'Hs.247813', 'Hs.247813', 'Hs.465643', 'Hs.352515'], 'Genome Version': ['February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)', 'February 2009 (Genome Reference Consortium GRCh37)'], 'Alignments': ['chr6:26271145-26271612 (-) // 100.0 // p22.2', 'chr6:26271145-26271612 (-) // 100.0 // p22.2', 'chr6:26271145-26271612 (-) // 100.0 // p22.2', 'chr19:4639529-5145579 (+) // 48.53 // p13.3', 'chr17:72920369-72929640 (+) // 100.0 // q25.1'], 'Gene Title': ['histone cluster 1, H3g', 'histone cluster 1, H3g', 'histone cluster 1, H3g', 'tumor necrosis factor, alpha-induced protein 8-like 1', 'otopetrin 2'], 'Gene Symbol': ['HIST1H3G', 'HIST1H3G', 'HIST1H3G', 'TNFAIP8L1', 'OTOP2'], 'Chromosomal Location': ['chr6p22.2', 'chr6p22.2', 'chr6p22.2', 'chr19p13.3', 'chr17q25.1'], 'Unigene Cluster Type': ['full length', 'full length', 'full length', 'full length', 'full length'], 'Ensembl': ['ENSG00000273983 /// OTTHUMG00000014436', 'ENSG00000273983 /// OTTHUMG00000014436', 'ENSG00000273983 /// OTTHUMG00000014436', 'ENSG00000185361 /// OTTHUMG00000182013', 'ENSG00000183034 /// OTTHUMG00000179215'], 'Entrez Gene': ['8355', '8355', '8355', '126282', '92736'], 'SwissProt': ['P68431', 'P68431', 'P68431', 'Q8WVP5', 'Q7RTS6'], 'EC': ['---', '---', '---', '---', '---'], 'OMIM': ['602815', '602815', '602815', '615869', '607827'], 'RefSeq Protein ID': ['NP_003525', 'NP_003525', 'NP_003525', 'NP_001161414 /// NP_689575 /// XP_005259544 /// XP_011525982', 'NP_835454 /// XP_011523781'], 'RefSeq Transcript ID': ['NM_003534', 'NM_003534', 'NM_003534', 'NM_001167942 /// NM_152362 /// XM_005259487 /// XM_011527680', 'NM_178160 /// XM_011525479'], 'Gene Ontology Biological Process': ['0000183 // chromatin silencing at rDNA // traceable author statement /// 0002230 // positive regulation of defense response to virus by host // inferred from mutant phenotype /// 0006325 // chromatin organization // traceable author statement /// 0006334 // nucleosome assembly // inferred from direct assay /// 0006334 // nucleosome assembly // inferred from mutant phenotype /// 0006335 // DNA replication-dependent nucleosome assembly // inferred from direct assay /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0010467 // gene expression // traceable author statement /// 0031047 // gene silencing by RNA // traceable author statement /// 0032776 // DNA methylation on cytosine // traceable author statement /// 0040029 // regulation of gene expression, epigenetic // traceable author statement /// 0044267 // cellular protein metabolic process // traceable author statement /// 0045814 // negative regulation of gene expression, epigenetic // traceable author statement /// 0051290 // protein heterotetramerization // inferred from direct assay /// 0060968 // regulation of gene silencing // inferred from direct assay /// 0098792 // xenophagy // inferred from mutant phenotype', '0000183 // chromatin silencing at rDNA // traceable author statement /// 0002230 // positive regulation of defense response to virus by host // inferred from mutant phenotype /// 0006325 // chromatin organization // traceable author statement /// 0006334 // nucleosome assembly // inferred from direct assay /// 0006334 // nucleosome assembly // inferred from mutant phenotype /// 0006335 // DNA replication-dependent nucleosome assembly // inferred from direct assay /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0010467 // gene expression // traceable author statement /// 0031047 // gene silencing by RNA // traceable author statement /// 0032776 // DNA methylation on cytosine // traceable author statement /// 0040029 // regulation of gene expression, epigenetic // traceable author statement /// 0044267 // cellular protein metabolic process // traceable author statement /// 0045814 // negative regulation of gene expression, epigenetic // traceable author statement /// 0051290 // protein heterotetramerization // inferred from direct assay /// 0060968 // regulation of gene silencing // inferred from direct assay /// 0098792 // xenophagy // inferred from mutant phenotype', '0000183 // chromatin silencing at rDNA // traceable author statement /// 0002230 // positive regulation of defense response to virus by host // inferred from mutant phenotype /// 0006325 // chromatin organization // traceable author statement /// 0006334 // nucleosome assembly // inferred from direct assay /// 0006334 // nucleosome assembly // inferred from mutant phenotype /// 0006335 // DNA replication-dependent nucleosome assembly // inferred from direct assay /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0010467 // gene expression // traceable author statement /// 0031047 // gene silencing by RNA // traceable author statement /// 0032776 // DNA methylation on cytosine // traceable author statement /// 0040029 // regulation of gene expression, epigenetic // traceable author statement /// 0044267 // cellular protein metabolic process // traceable author statement /// 0045814 // negative regulation of gene expression, epigenetic // traceable author statement /// 0051290 // protein heterotetramerization // inferred from direct assay /// 0060968 // regulation of gene silencing // inferred from direct assay /// 0098792 // xenophagy // inferred from mutant phenotype', '0032007 // negative regulation of TOR signaling // not recorded /// 0032007 // negative regulation of TOR signaling // inferred from sequence or structural similarity', '---'], 'Gene Ontology Cellular Component': ['0000228 // nuclear chromosome // inferred from direct assay /// 0000786 // nucleosome // inferred from direct assay /// 0000788 // nuclear nucleosome // inferred from direct assay /// 0005576 // extracellular region // traceable author statement /// 0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // traceable author statement /// 0005694 // chromosome // inferred from electronic annotation /// 0016020 // membrane // inferred from direct assay /// 0043234 // protein complex // inferred from direct assay /// 0070062 // extracellular exosome // inferred from direct assay', '0000228 // nuclear chromosome // inferred from direct assay /// 0000786 // nucleosome // inferred from direct assay /// 0000788 // nuclear nucleosome // inferred from direct assay /// 0005576 // extracellular region // traceable author statement /// 0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // traceable author statement /// 0005694 // chromosome // inferred from electronic annotation /// 0016020 // membrane // inferred from direct assay /// 0043234 // protein complex // inferred from direct assay /// 0070062 // extracellular exosome // inferred from direct assay', '0000228 // nuclear chromosome // inferred from direct assay /// 0000786 // nucleosome // inferred from direct assay /// 0000788 // nuclear nucleosome // inferred from direct assay /// 0005576 // extracellular region // traceable author statement /// 0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // traceable author statement /// 0005694 // chromosome // inferred from electronic annotation /// 0016020 // membrane // inferred from direct assay /// 0043234 // protein complex // inferred from direct assay /// 0070062 // extracellular exosome // inferred from direct assay', '0005737 // cytoplasm // not recorded /// 0005737 // cytoplasm // inferred from sequence or structural similarity', '0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation'], 'Gene Ontology Molecular Function': ['0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0042393 // histone binding // inferred from physical interaction /// 0046982 // protein heterodimerization activity // inferred from electronic annotation', '0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0042393 // histone binding // inferred from physical interaction /// 0046982 // protein heterodimerization activity // inferred from electronic annotation', '0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0042393 // histone binding // inferred from physical interaction /// 0046982 // protein heterodimerization activity // inferred from electronic annotation', '0005515 // protein binding // inferred from physical interaction', '---'], 'Pathway': ['---', '---', '---', '---', '---'], 'InterPro': ['IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 /// IPR007125 // Histone H2A/H2B/H3 // 1.7E-37', 'IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 /// IPR007125 // Histone H2A/H2B/H3 // 1.7E-37', 'IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 /// IPR007125 // Histone H2A/H2B/H3 // 1.7E-37', 'IPR008477 // Protein of unknown function DUF758 // 8.4E-86 /// IPR008477 // Protein of unknown function DUF758 // 6.8E-90', 'IPR004878 // Otopetrin // 9.4E-43 /// IPR004878 // Otopetrin // 9.4E-43 /// IPR004878 // Otopetrin // 9.4E-43 /// IPR004878 // Otopetrin // 3.9E-18 /// IPR004878 // Otopetrin // 3.8E-20 /// IPR004878 // Otopetrin // 5.2E-16'], 'Annotation Description': ['This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 4 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 4 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 4 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 9 transcripts. // false // Matching Probes // A', 'This probe set was annotated using the Matching Probes based pipeline to a Entrez Gene identifier using 6 transcripts. // false // Matching Probes // A'], 'Annotation Transcript Cluster': ['ENST00000614378(11),NM_003534(11),OTTHUMT00000040099(11),uc003nhi.3', 'ENST00000614378(11),NM_003534(11),OTTHUMT00000040099(11),uc003nhi.3', 'ENST00000614378(11),NM_003534(11),OTTHUMT00000040099(11),uc003nhi.3', 'BC017672(11),BC044250(9),ENST00000327473(11),ENST00000536716(11),NM_001167942(11),NM_152362(11),OTTHUMT00000458662(11),uc002max.3,uc021une.1', 'ENST00000331427(11),ENST00000580223(11),NM_178160(11),OTTHUMT00000445306(11),uc010wrp.2,XM_011525479(11)'], 'Transcript Assignments': ['ENST00000614378 // ensembl_havana_transcript:known chromosome:GRCh38:6:26269405:26271815:-1 gene:ENSG00000273983 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_003534 // Homo sapiens histone cluster 1, H3g (HIST1H3G), mRNA. // refseq // 11 // --- /// OTTHUMT00000040099 // otter:known chromosome:VEGA61:6:26269405:26271815:-1 gene:OTTHUMG00000014436 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc003nhi.3 // --- // ucsc_genes // 11 // ---', 'ENST00000614378 // ensembl_havana_transcript:known chromosome:GRCh38:6:26269405:26271815:-1 gene:ENSG00000273983 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// GENSCAN00000029819 // cdna:genscan chromosome:GRCh38:6:26270974:26271384:-1 transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_003534 // Homo sapiens histone cluster 1, H3g (HIST1H3G), mRNA. // refseq // 11 // --- /// OTTHUMT00000040099 // otter:known chromosome:VEGA61:6:26269405:26271815:-1 gene:OTTHUMG00000014436 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc003nhi.3 // --- // ucsc_genes // 11 // ---', 'ENST00000614378 // ensembl_havana_transcript:known chromosome:GRCh38:6:26269405:26271815:-1 gene:ENSG00000273983 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_003534 // Homo sapiens histone cluster 1, H3g (HIST1H3G), mRNA. // refseq // 11 // --- /// OTTHUMT00000040099 // otter:known chromosome:VEGA61:6:26269405:26271815:-1 gene:OTTHUMG00000014436 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc003nhi.3 // --- // ucsc_genes // 11 // ---', 'BC017672 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1, mRNA (cDNA clone MGC:17791 IMAGE:3885999), complete cds. // gb // 11 // --- /// BC044250 // accn=BC044250 class=mRNAlike lncRNA name=Human lncRNA ref=JounralRNA transcriptId=673 cpcScore=-0.1526100 cnci=-0.1238602 // noncode // 9 // --- /// BC044250 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1, mRNA (cDNA clone IMAGE:5784807). // gb // 9 // --- /// ENST00000327473 // ensembl_havana_transcript:known chromosome:GRCh38:19:4639518:4655568:1 gene:ENSG00000185361 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000536716 // ensembl:known chromosome:GRCh38:19:4640017:4655568:1 gene:ENSG00000185361 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_001167942 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1 (TNFAIP8L1), transcript variant 1, mRNA. // refseq // 11 // --- /// NM_152362 // Homo sapiens tumor necrosis factor, alpha-induced protein 8-like 1 (TNFAIP8L1), transcript variant 2, mRNA. // refseq // 11 // --- /// NONHSAT060631 // Non-coding transcript identified by NONCODE: Exonic // noncode // 9 // --- /// OTTHUMT00000458662 // otter:known chromosome:VEGA61:19:4639518:4655568:1 gene:OTTHUMG00000182013 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc002max.3 // --- // ucsc_genes // 11 // --- /// uc021une.1 // --- // ucsc_genes // 11 // ---', 'ENST00000331427 // ensembl:known chromosome:GRCh38:17:74924275:74933911:1 gene:ENSG00000183034 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// ENST00000580223 // havana:known chromosome:GRCh38:17:74924603:74933912:1 gene:ENSG00000183034 gene_biotype:protein_coding transcript_biotype:protein_coding // ensembl // 11 // --- /// GENSCAN00000013715 // cdna:genscan chromosome:GRCh38:17:74924633:74933545:1 transcript_biotype:protein_coding // ensembl // 11 // --- /// NM_178160 // Homo sapiens otopetrin 2 (OTOP2), mRNA. // refseq // 11 // --- /// OTTHUMT00000445306 // otter:known chromosome:VEGA61:17:74924603:74933912:1 gene:OTTHUMG00000179215 gene_biotype:protein_coding transcript_biotype:protein_coding // vega // 11 // --- /// uc010wrp.2 // --- // ucsc_genes // 11 // --- /// XM_011525479 // PREDICTED: Homo sapiens otopetrin 2 (OTOP2), transcript variant X1, mRNA. // refseq // 11 // ---'], 'Annotation Notes': ['---', '---', 'GENSCAN00000029819 // ensembl // 4 // Cross Hyb Matching Probes', '---', '---'], 'SPOT_ID': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GeneChip Array</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Transcript ID(Array Design)</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>GI</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "      <th>Pathway</th>\n",
       "      <th>InterPro</th>\n",
       "      <th>Annotation Description</th>\n",
       "      <th>Annotation Transcript Cluster</th>\n",
       "      <th>Transcript Assignments</th>\n",
       "      <th>Annotation Notes</th>\n",
       "      <th>SPOT_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11715100_at</td>\n",
       "      <td>Human Genome PrimeView Array</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>30-Mar-16</td>\n",
       "      <td>Consensus sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>g21264570</td>\n",
       "      <td>g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /T...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21264570.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0000183 // chromatin silencing at rDNA // trac...</td>\n",
       "      <td>0000228 // nuclear chromosome // inferred from...</td>\n",
       "      <td>0003677 // DNA binding // inferred from electr...</td>\n",
       "      <td>---</td>\n",
       "      <td>IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 ///...</td>\n",
       "      <td>This probe set was annotated using the Matchin...</td>\n",
       "      <td>ENST00000614378(11),NM_003534(11),OTTHUMT00000...</td>\n",
       "      <td>ENST00000614378 // ensembl_havana_transcript:k...</td>\n",
       "      <td>---</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11715101_s_at</td>\n",
       "      <td>Human Genome PrimeView Array</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>30-Mar-16</td>\n",
       "      <td>Consensus sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>g21264570</td>\n",
       "      <td>g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /T...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21264570.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0000183 // chromatin silencing at rDNA // trac...</td>\n",
       "      <td>0000228 // nuclear chromosome // inferred from...</td>\n",
       "      <td>0003677 // DNA binding // inferred from electr...</td>\n",
       "      <td>---</td>\n",
       "      <td>IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 ///...</td>\n",
       "      <td>This probe set was annotated using the Matchin...</td>\n",
       "      <td>ENST00000614378(11),NM_003534(11),OTTHUMT00000...</td>\n",
       "      <td>ENST00000614378 // ensembl_havana_transcript:k...</td>\n",
       "      <td>---</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11715102_x_at</td>\n",
       "      <td>Human Genome PrimeView Array</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>30-Mar-16</td>\n",
       "      <td>Consensus sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>g21264570</td>\n",
       "      <td>g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /T...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21264570.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0000183 // chromatin silencing at rDNA // trac...</td>\n",
       "      <td>0000228 // nuclear chromosome // inferred from...</td>\n",
       "      <td>0003677 // DNA binding // inferred from electr...</td>\n",
       "      <td>---</td>\n",
       "      <td>IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 ///...</td>\n",
       "      <td>This probe set was annotated using the Matchin...</td>\n",
       "      <td>ENST00000614378(11),NM_003534(11),OTTHUMT00000...</td>\n",
       "      <td>ENST00000614378 // ensembl_havana_transcript:k...</td>\n",
       "      <td>GENSCAN00000029819 // ensembl // 4 // Cross Hy...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11715103_x_at</td>\n",
       "      <td>Human Genome PrimeView Array</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>30-Mar-16</td>\n",
       "      <td>Consensus sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>g22748780</td>\n",
       "      <td>g22748780 /TID=g22748780 /CNT=1 /FEA=FLmRNA /T...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22748780.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0032007 // negative regulation of TOR signalin...</td>\n",
       "      <td>0005737 // cytoplasm // not recorded /// 00057...</td>\n",
       "      <td>0005515 // protein binding // inferred from ph...</td>\n",
       "      <td>---</td>\n",
       "      <td>IPR008477 // Protein of unknown function DUF75...</td>\n",
       "      <td>This probe set was annotated using the Matchin...</td>\n",
       "      <td>BC017672(11),BC044250(9),ENST00000327473(11),E...</td>\n",
       "      <td>BC017672 // Homo sapiens tumor necrosis factor...</td>\n",
       "      <td>---</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11715104_s_at</td>\n",
       "      <td>Human Genome PrimeView Array</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>30-Mar-16</td>\n",
       "      <td>Consensus sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>g30039713</td>\n",
       "      <td>g30039713 /TID=g30039713 /CNT=1 /FEA=FLmRNA /T...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30039713.0</td>\n",
       "      <td>...</td>\n",
       "      <td>---</td>\n",
       "      <td>0016020 // membrane // inferred from electroni...</td>\n",
       "      <td>---</td>\n",
       "      <td>---</td>\n",
       "      <td>IPR004878 // Otopetrin // 9.4E-43 /// IPR00487...</td>\n",
       "      <td>This probe set was annotated using the Matchin...</td>\n",
       "      <td>ENST00000331427(11),ENST00000580223(11),NM_178...</td>\n",
       "      <td>ENST00000331427 // ensembl:known chromosome:GR...</td>\n",
       "      <td>---</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161338</th>\n",
       "      <td>AFFX-r2-TagJ-5_at</td>\n",
       "      <td>4.615234375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161339</th>\n",
       "      <td>AFFX-r2-TagO-3_at</td>\n",
       "      <td>4.705078125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161340</th>\n",
       "      <td>AFFX-r2-TagO-5_at</td>\n",
       "      <td>4.608398438</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161341</th>\n",
       "      <td>AFFX-r2-TagQ-3_at</td>\n",
       "      <td>4.697265625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161342</th>\n",
       "      <td>AFFX-r2-TagQ-5_at</td>\n",
       "      <td>4.671875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3161343 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        ID                GeneChip Array  \\\n",
       "0              11715100_at  Human Genome PrimeView Array   \n",
       "1            11715101_s_at  Human Genome PrimeView Array   \n",
       "2            11715102_x_at  Human Genome PrimeView Array   \n",
       "3            11715103_x_at  Human Genome PrimeView Array   \n",
       "4            11715104_s_at  Human Genome PrimeView Array   \n",
       "...                    ...                           ...   \n",
       "3161338  AFFX-r2-TagJ-5_at                   4.615234375   \n",
       "3161339  AFFX-r2-TagO-3_at                   4.705078125   \n",
       "3161340  AFFX-r2-TagO-5_at                   4.608398438   \n",
       "3161341  AFFX-r2-TagQ-3_at                   4.697265625   \n",
       "3161342  AFFX-r2-TagQ-5_at                      4.671875   \n",
       "\n",
       "        Species Scientific Name Annotation Date       Sequence Type  \\\n",
       "0                  Homo sapiens       30-Mar-16  Consensus sequence   \n",
       "1                  Homo sapiens       30-Mar-16  Consensus sequence   \n",
       "2                  Homo sapiens       30-Mar-16  Consensus sequence   \n",
       "3                  Homo sapiens       30-Mar-16  Consensus sequence   \n",
       "4                  Homo sapiens       30-Mar-16  Consensus sequence   \n",
       "...                         ...             ...                 ...   \n",
       "3161338                     NaN             NaN                 NaN   \n",
       "3161339                     NaN             NaN                 NaN   \n",
       "3161340                     NaN             NaN                 NaN   \n",
       "3161341                     NaN             NaN                 NaN   \n",
       "3161342                     NaN             NaN                 NaN   \n",
       "\n",
       "                         Sequence Source Transcript ID(Array Design)  \\\n",
       "0        Affymetrix Proprietary Database                   g21264570   \n",
       "1        Affymetrix Proprietary Database                   g21264570   \n",
       "2        Affymetrix Proprietary Database                   g21264570   \n",
       "3        Affymetrix Proprietary Database                   g22748780   \n",
       "4        Affymetrix Proprietary Database                   g30039713   \n",
       "...                                  ...                         ...   \n",
       "3161338                              NaN                         NaN   \n",
       "3161339                              NaN                         NaN   \n",
       "3161340                              NaN                         NaN   \n",
       "3161341                              NaN                         NaN   \n",
       "3161342                              NaN                         NaN   \n",
       "\n",
       "                                        Target Description GB_ACC          GI  \\\n",
       "0        g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /T...    NaN  21264570.0   \n",
       "1        g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /T...    NaN  21264570.0   \n",
       "2        g21264570 /TID=g21264570 /CNT=1 /FEA=FLmRNA /T...    NaN  21264570.0   \n",
       "3        g22748780 /TID=g22748780 /CNT=1 /FEA=FLmRNA /T...    NaN  22748780.0   \n",
       "4        g30039713 /TID=g30039713 /CNT=1 /FEA=FLmRNA /T...    NaN  30039713.0   \n",
       "...                                                    ...    ...         ...   \n",
       "3161338                                                NaN    NaN         NaN   \n",
       "3161339                                                NaN    NaN         NaN   \n",
       "3161340                                                NaN    NaN         NaN   \n",
       "3161341                                                NaN    NaN         NaN   \n",
       "3161342                                                NaN    NaN         NaN   \n",
       "\n",
       "         ...                   Gene Ontology Biological Process  \\\n",
       "0        ...  0000183 // chromatin silencing at rDNA // trac...   \n",
       "1        ...  0000183 // chromatin silencing at rDNA // trac...   \n",
       "2        ...  0000183 // chromatin silencing at rDNA // trac...   \n",
       "3        ...  0032007 // negative regulation of TOR signalin...   \n",
       "4        ...                                                ---   \n",
       "...      ...                                                ...   \n",
       "3161338  ...                                                NaN   \n",
       "3161339  ...                                                NaN   \n",
       "3161340  ...                                                NaN   \n",
       "3161341  ...                                                NaN   \n",
       "3161342  ...                                                NaN   \n",
       "\n",
       "                          Gene Ontology Cellular Component  \\\n",
       "0        0000228 // nuclear chromosome // inferred from...   \n",
       "1        0000228 // nuclear chromosome // inferred from...   \n",
       "2        0000228 // nuclear chromosome // inferred from...   \n",
       "3        0005737 // cytoplasm // not recorded /// 00057...   \n",
       "4        0016020 // membrane // inferred from electroni...   \n",
       "...                                                    ...   \n",
       "3161338                                                NaN   \n",
       "3161339                                                NaN   \n",
       "3161340                                                NaN   \n",
       "3161341                                                NaN   \n",
       "3161342                                                NaN   \n",
       "\n",
       "                          Gene Ontology Molecular Function Pathway  \\\n",
       "0        0003677 // DNA binding // inferred from electr...     ---   \n",
       "1        0003677 // DNA binding // inferred from electr...     ---   \n",
       "2        0003677 // DNA binding // inferred from electr...     ---   \n",
       "3        0005515 // protein binding // inferred from ph...     ---   \n",
       "4                                                      ---     ---   \n",
       "...                                                    ...     ...   \n",
       "3161338                                                NaN     NaN   \n",
       "3161339                                                NaN     NaN   \n",
       "3161340                                                NaN     NaN   \n",
       "3161341                                                NaN     NaN   \n",
       "3161342                                                NaN     NaN   \n",
       "\n",
       "                                                  InterPro  \\\n",
       "0        IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 ///...   \n",
       "1        IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 ///...   \n",
       "2        IPR007125 // Histone H2A/H2B/H3 // 9.3E-34 ///...   \n",
       "3        IPR008477 // Protein of unknown function DUF75...   \n",
       "4        IPR004878 // Otopetrin // 9.4E-43 /// IPR00487...   \n",
       "...                                                    ...   \n",
       "3161338                                                NaN   \n",
       "3161339                                                NaN   \n",
       "3161340                                                NaN   \n",
       "3161341                                                NaN   \n",
       "3161342                                                NaN   \n",
       "\n",
       "                                    Annotation Description  \\\n",
       "0        This probe set was annotated using the Matchin...   \n",
       "1        This probe set was annotated using the Matchin...   \n",
       "2        This probe set was annotated using the Matchin...   \n",
       "3        This probe set was annotated using the Matchin...   \n",
       "4        This probe set was annotated using the Matchin...   \n",
       "...                                                    ...   \n",
       "3161338                                                NaN   \n",
       "3161339                                                NaN   \n",
       "3161340                                                NaN   \n",
       "3161341                                                NaN   \n",
       "3161342                                                NaN   \n",
       "\n",
       "                             Annotation Transcript Cluster  \\\n",
       "0        ENST00000614378(11),NM_003534(11),OTTHUMT00000...   \n",
       "1        ENST00000614378(11),NM_003534(11),OTTHUMT00000...   \n",
       "2        ENST00000614378(11),NM_003534(11),OTTHUMT00000...   \n",
       "3        BC017672(11),BC044250(9),ENST00000327473(11),E...   \n",
       "4        ENST00000331427(11),ENST00000580223(11),NM_178...   \n",
       "...                                                    ...   \n",
       "3161338                                                NaN   \n",
       "3161339                                                NaN   \n",
       "3161340                                                NaN   \n",
       "3161341                                                NaN   \n",
       "3161342                                                NaN   \n",
       "\n",
       "                                    Transcript Assignments  \\\n",
       "0        ENST00000614378 // ensembl_havana_transcript:k...   \n",
       "1        ENST00000614378 // ensembl_havana_transcript:k...   \n",
       "2        ENST00000614378 // ensembl_havana_transcript:k...   \n",
       "3        BC017672 // Homo sapiens tumor necrosis factor...   \n",
       "4        ENST00000331427 // ensembl:known chromosome:GR...   \n",
       "...                                                    ...   \n",
       "3161338                                                NaN   \n",
       "3161339                                                NaN   \n",
       "3161340                                                NaN   \n",
       "3161341                                                NaN   \n",
       "3161342                                                NaN   \n",
       "\n",
       "                                          Annotation Notes SPOT_ID  \n",
       "0                                                      ---     NaN  \n",
       "1                                                      ---     NaN  \n",
       "2        GENSCAN00000029819 // ensembl // 4 // Cross Hy...     NaN  \n",
       "3                                                      ---     NaN  \n",
       "4                                                      ---     NaN  \n",
       "...                                                    ...     ...  \n",
       "3161338                                                NaN     NaN  \n",
       "3161339                                                NaN     NaN  \n",
       "3161340                                                NaN     NaN  \n",
       "3161341                                                NaN     NaN  \n",
       "3161342                                                NaN     NaN  \n",
       "\n",
       "[3161343 rows x 36 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GeneChip Array', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Transcript ID(Array Design)',\n",
       "       'Target Description', 'GB_ACC', 'GI', 'Representative Public ID',\n",
       "       'Archival UniGene Cluster', 'UniGene ID', 'Genome Version',\n",
       "       'Alignments', 'Gene Title', 'Gene Symbol', 'Chromosomal Location',\n",
       "       'Unigene Cluster Type', 'Ensembl', 'Entrez Gene', 'SwissProt', 'EC',\n",
       "       'OMIM', 'RefSeq Protein ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function', 'Pathway', 'InterPro',\n",
       "       'Annotation Description', 'Annotation Transcript Cluster',\n",
       "       'Transcript Assignments', 'Annotation Notes', 'SPOT_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2411058</th>\n",
       "      <th>GSM2411059</th>\n",
       "      <th>GSM2411060</th>\n",
       "      <th>GSM2411061</th>\n",
       "      <th>GSM2411062</th>\n",
       "      <th>GSM2411063</th>\n",
       "      <th>GSM2411064</th>\n",
       "      <th>GSM2411065</th>\n",
       "      <th>GSM2411066</th>\n",
       "      <th>GSM2411067</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2411111</th>\n",
       "      <th>GSM2411112</th>\n",
       "      <th>GSM2411113</th>\n",
       "      <th>GSM2411114</th>\n",
       "      <th>GSM2411115</th>\n",
       "      <th>GSM2411116</th>\n",
       "      <th>GSM2411117</th>\n",
       "      <th>GSM2411118</th>\n",
       "      <th>GSM2411119</th>\n",
       "      <th>GSM2411120</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>5.180866</td>\n",
       "      <td>4.977051</td>\n",
       "      <td>5.195313</td>\n",
       "      <td>4.862793</td>\n",
       "      <td>4.852539</td>\n",
       "      <td>5.636719</td>\n",
       "      <td>6.330566</td>\n",
       "      <td>5.047852</td>\n",
       "      <td>5.034180</td>\n",
       "      <td>5.624512</td>\n",
       "      <td>...</td>\n",
       "      <td>4.866211</td>\n",
       "      <td>4.717285</td>\n",
       "      <td>5.069336</td>\n",
       "      <td>4.842285</td>\n",
       "      <td>4.742676</td>\n",
       "      <td>4.809570</td>\n",
       "      <td>4.792969</td>\n",
       "      <td>5.490723</td>\n",
       "      <td>5.702148</td>\n",
       "      <td>6.715332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>5.471013</td>\n",
       "      <td>5.002930</td>\n",
       "      <td>5.356445</td>\n",
       "      <td>4.798828</td>\n",
       "      <td>4.750977</td>\n",
       "      <td>6.299805</td>\n",
       "      <td>7.727539</td>\n",
       "      <td>5.214844</td>\n",
       "      <td>5.150391</td>\n",
       "      <td>6.301758</td>\n",
       "      <td>...</td>\n",
       "      <td>4.853516</td>\n",
       "      <td>4.749023</td>\n",
       "      <td>5.251953</td>\n",
       "      <td>4.809570</td>\n",
       "      <td>4.689453</td>\n",
       "      <td>4.787109</td>\n",
       "      <td>4.773438</td>\n",
       "      <td>6.125977</td>\n",
       "      <td>6.534180</td>\n",
       "      <td>8.548828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOT2</th>\n",
       "      <td>7.713635</td>\n",
       "      <td>6.903320</td>\n",
       "      <td>7.163086</td>\n",
       "      <td>8.966797</td>\n",
       "      <td>8.912109</td>\n",
       "      <td>7.658203</td>\n",
       "      <td>7.423828</td>\n",
       "      <td>6.889648</td>\n",
       "      <td>7.335938</td>\n",
       "      <td>7.115234</td>\n",
       "      <td>...</td>\n",
       "      <td>7.334961</td>\n",
       "      <td>7.650391</td>\n",
       "      <td>8.146484</td>\n",
       "      <td>7.863281</td>\n",
       "      <td>7.859375</td>\n",
       "      <td>7.816406</td>\n",
       "      <td>8.263672</td>\n",
       "      <td>5.544922</td>\n",
       "      <td>6.828125</td>\n",
       "      <td>7.619141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM2B</th>\n",
       "      <td>4.743009</td>\n",
       "      <td>4.749023</td>\n",
       "      <td>4.800000</td>\n",
       "      <td>4.703711</td>\n",
       "      <td>4.716211</td>\n",
       "      <td>4.777148</td>\n",
       "      <td>4.722656</td>\n",
       "      <td>4.741016</td>\n",
       "      <td>4.869922</td>\n",
       "      <td>4.773828</td>\n",
       "      <td>...</td>\n",
       "      <td>4.690039</td>\n",
       "      <td>4.709961</td>\n",
       "      <td>4.745898</td>\n",
       "      <td>4.784766</td>\n",
       "      <td>4.738672</td>\n",
       "      <td>4.766797</td>\n",
       "      <td>4.747461</td>\n",
       "      <td>4.816797</td>\n",
       "      <td>4.822461</td>\n",
       "      <td>4.763086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACY1</th>\n",
       "      <td>6.272151</td>\n",
       "      <td>6.059570</td>\n",
       "      <td>6.395020</td>\n",
       "      <td>7.125488</td>\n",
       "      <td>7.118164</td>\n",
       "      <td>6.563965</td>\n",
       "      <td>6.569336</td>\n",
       "      <td>7.001953</td>\n",
       "      <td>6.913086</td>\n",
       "      <td>7.670410</td>\n",
       "      <td>...</td>\n",
       "      <td>5.946777</td>\n",
       "      <td>6.174805</td>\n",
       "      <td>6.546875</td>\n",
       "      <td>5.881836</td>\n",
       "      <td>6.570313</td>\n",
       "      <td>6.386230</td>\n",
       "      <td>5.936035</td>\n",
       "      <td>6.612793</td>\n",
       "      <td>7.480957</td>\n",
       "      <td>6.257324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>4.680315</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>4.670898</td>\n",
       "      <td>4.657715</td>\n",
       "      <td>4.712891</td>\n",
       "      <td>4.703125</td>\n",
       "      <td>4.690918</td>\n",
       "      <td>4.664551</td>\n",
       "      <td>4.665039</td>\n",
       "      <td>4.683594</td>\n",
       "      <td>...</td>\n",
       "      <td>4.670410</td>\n",
       "      <td>4.671387</td>\n",
       "      <td>4.714355</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>4.674316</td>\n",
       "      <td>4.694336</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>4.666992</td>\n",
       "      <td>4.657227</td>\n",
       "      <td>4.693359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>6.612026</td>\n",
       "      <td>6.639160</td>\n",
       "      <td>6.410156</td>\n",
       "      <td>6.420166</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>6.626953</td>\n",
       "      <td>6.472412</td>\n",
       "      <td>6.822266</td>\n",
       "      <td>6.800537</td>\n",
       "      <td>6.394043</td>\n",
       "      <td>...</td>\n",
       "      <td>6.310547</td>\n",
       "      <td>6.310059</td>\n",
       "      <td>6.320801</td>\n",
       "      <td>6.378174</td>\n",
       "      <td>6.380615</td>\n",
       "      <td>6.425537</td>\n",
       "      <td>6.324707</td>\n",
       "      <td>6.276855</td>\n",
       "      <td>6.554199</td>\n",
       "      <td>6.743652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>6.074064</td>\n",
       "      <td>7.055664</td>\n",
       "      <td>6.838867</td>\n",
       "      <td>6.343750</td>\n",
       "      <td>6.431152</td>\n",
       "      <td>6.651855</td>\n",
       "      <td>6.940430</td>\n",
       "      <td>6.520508</td>\n",
       "      <td>7.701660</td>\n",
       "      <td>6.390137</td>\n",
       "      <td>...</td>\n",
       "      <td>6.027832</td>\n",
       "      <td>6.108398</td>\n",
       "      <td>7.544434</td>\n",
       "      <td>6.250977</td>\n",
       "      <td>6.095703</td>\n",
       "      <td>6.333496</td>\n",
       "      <td>6.608398</td>\n",
       "      <td>6.477051</td>\n",
       "      <td>6.934570</td>\n",
       "      <td>6.090332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>5.904870</td>\n",
       "      <td>5.963867</td>\n",
       "      <td>6.434082</td>\n",
       "      <td>5.895508</td>\n",
       "      <td>5.866699</td>\n",
       "      <td>6.193359</td>\n",
       "      <td>5.788086</td>\n",
       "      <td>6.060547</td>\n",
       "      <td>5.915039</td>\n",
       "      <td>6.078613</td>\n",
       "      <td>...</td>\n",
       "      <td>6.214355</td>\n",
       "      <td>6.235352</td>\n",
       "      <td>6.056152</td>\n",
       "      <td>6.332520</td>\n",
       "      <td>5.993164</td>\n",
       "      <td>6.766602</td>\n",
       "      <td>5.933594</td>\n",
       "      <td>6.520996</td>\n",
       "      <td>6.223633</td>\n",
       "      <td>5.875000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>7.978461</td>\n",
       "      <td>7.258789</td>\n",
       "      <td>7.264160</td>\n",
       "      <td>7.532715</td>\n",
       "      <td>7.452637</td>\n",
       "      <td>7.616211</td>\n",
       "      <td>7.958984</td>\n",
       "      <td>8.042480</td>\n",
       "      <td>6.781250</td>\n",
       "      <td>8.026855</td>\n",
       "      <td>...</td>\n",
       "      <td>7.327148</td>\n",
       "      <td>7.410156</td>\n",
       "      <td>6.974121</td>\n",
       "      <td>7.502930</td>\n",
       "      <td>7.506348</td>\n",
       "      <td>7.829102</td>\n",
       "      <td>6.866699</td>\n",
       "      <td>6.866211</td>\n",
       "      <td>7.709961</td>\n",
       "      <td>7.666992</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21758 rows Ã— 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM2411058  GSM2411059  GSM2411060  GSM2411061  GSM2411062  \\\n",
       "Gene                                                                    \n",
       " ABCC6P1     5.180866    4.977051    5.195313    4.862793    4.852539   \n",
       " ABCC6P2     5.471013    5.002930    5.356445    4.798828    4.750977   \n",
       " ACOT2       7.713635    6.903320    7.163086    8.966797    8.912109   \n",
       " ACSM2B      4.743009    4.749023    4.800000    4.703711    4.716211   \n",
       " ACY1        6.272151    6.059570    6.395020    7.125488    7.118164   \n",
       "...               ...         ...         ...         ...         ...   \n",
       "ZYG11A       4.680315    4.662598    4.670898    4.657715    4.712891   \n",
       "ZYG11B       6.612026    6.639160    6.410156    6.420166    6.500000   \n",
       "ZYX          6.074064    7.055664    6.838867    6.343750    6.431152   \n",
       "ZZEF1        5.904870    5.963867    6.434082    5.895508    5.866699   \n",
       "ZZZ3         7.978461    7.258789    7.264160    7.532715    7.452637   \n",
       "\n",
       "           GSM2411063  GSM2411064  GSM2411065  GSM2411066  GSM2411067  ...  \\\n",
       "Gene                                                                   ...   \n",
       " ABCC6P1     5.636719    6.330566    5.047852    5.034180    5.624512  ...   \n",
       " ABCC6P2     6.299805    7.727539    5.214844    5.150391    6.301758  ...   \n",
       " ACOT2       7.658203    7.423828    6.889648    7.335938    7.115234  ...   \n",
       " ACSM2B      4.777148    4.722656    4.741016    4.869922    4.773828  ...   \n",
       " ACY1        6.563965    6.569336    7.001953    6.913086    7.670410  ...   \n",
       "...               ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A       4.703125    4.690918    4.664551    4.665039    4.683594  ...   \n",
       "ZYG11B       6.626953    6.472412    6.822266    6.800537    6.394043  ...   \n",
       "ZYX          6.651855    6.940430    6.520508    7.701660    6.390137  ...   \n",
       "ZZEF1        6.193359    5.788086    6.060547    5.915039    6.078613  ...   \n",
       "ZZZ3         7.616211    7.958984    8.042480    6.781250    8.026855  ...   \n",
       "\n",
       "           GSM2411111  GSM2411112  GSM2411113  GSM2411114  GSM2411115  \\\n",
       "Gene                                                                    \n",
       " ABCC6P1     4.866211    4.717285    5.069336    4.842285    4.742676   \n",
       " ABCC6P2     4.853516    4.749023    5.251953    4.809570    4.689453   \n",
       " ACOT2       7.334961    7.650391    8.146484    7.863281    7.859375   \n",
       " ACSM2B      4.690039    4.709961    4.745898    4.784766    4.738672   \n",
       " ACY1        5.946777    6.174805    6.546875    5.881836    6.570313   \n",
       "...               ...         ...         ...         ...         ...   \n",
       "ZYG11A       4.670410    4.671387    4.714355    4.662598    4.674316   \n",
       "ZYG11B       6.310547    6.310059    6.320801    6.378174    6.380615   \n",
       "ZYX          6.027832    6.108398    7.544434    6.250977    6.095703   \n",
       "ZZEF1        6.214355    6.235352    6.056152    6.332520    5.993164   \n",
       "ZZZ3         7.327148    7.410156    6.974121    7.502930    7.506348   \n",
       "\n",
       "           GSM2411116  GSM2411117  GSM2411118  GSM2411119  GSM2411120  \n",
       "Gene                                                                   \n",
       " ABCC6P1     4.809570    4.792969    5.490723    5.702148    6.715332  \n",
       " ABCC6P2     4.787109    4.773438    6.125977    6.534180    8.548828  \n",
       " ACOT2       7.816406    8.263672    5.544922    6.828125    7.619141  \n",
       " ACSM2B      4.766797    4.747461    4.816797    4.822461    4.763086  \n",
       " ACY1        6.386230    5.936035    6.612793    7.480957    6.257324  \n",
       "...               ...         ...         ...         ...         ...  \n",
       "ZYG11A       4.694336    4.662598    4.666992    4.657227    4.693359  \n",
       "ZYG11B       6.425537    6.324707    6.276855    6.554199    6.743652  \n",
       "ZYX          6.333496    6.608398    6.477051    6.934570    6.090332  \n",
       "ZZEF1        6.766602    5.933594    6.520996    6.223633    5.875000  \n",
       "ZZZ3         7.829102    6.866699    6.866211    7.709961    7.666992  \n",
       "\n",
       "[21758 rows x 63 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2411058</th>\n",
       "      <th>GSM2411059</th>\n",
       "      <th>GSM2411060</th>\n",
       "      <th>GSM2411061</th>\n",
       "      <th>GSM2411062</th>\n",
       "      <th>GSM2411063</th>\n",
       "      <th>GSM2411064</th>\n",
       "      <th>GSM2411065</th>\n",
       "      <th>GSM2411066</th>\n",
       "      <th>GSM2411067</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2411111</th>\n",
       "      <th>GSM2411112</th>\n",
       "      <th>GSM2411113</th>\n",
       "      <th>GSM2411114</th>\n",
       "      <th>GSM2411115</th>\n",
       "      <th>GSM2411116</th>\n",
       "      <th>GSM2411117</th>\n",
       "      <th>GSM2411118</th>\n",
       "      <th>GSM2411119</th>\n",
       "      <th>GSM2411120</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.106244</td>\n",
       "      <td>5.238281</td>\n",
       "      <td>5.280273</td>\n",
       "      <td>5.072266</td>\n",
       "      <td>5.064453</td>\n",
       "      <td>5.056641</td>\n",
       "      <td>6.063477</td>\n",
       "      <td>4.973633</td>\n",
       "      <td>5.327148</td>\n",
       "      <td>5.094727</td>\n",
       "      <td>...</td>\n",
       "      <td>5.244141</td>\n",
       "      <td>4.891602</td>\n",
       "      <td>5.267578</td>\n",
       "      <td>5.543945</td>\n",
       "      <td>5.151367</td>\n",
       "      <td>5.382812</td>\n",
       "      <td>5.319336</td>\n",
       "      <td>4.903320</td>\n",
       "      <td>5.154297</td>\n",
       "      <td>6.026367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>4.872861</td>\n",
       "      <td>4.875000</td>\n",
       "      <td>4.913086</td>\n",
       "      <td>4.846191</td>\n",
       "      <td>4.951660</td>\n",
       "      <td>4.916016</td>\n",
       "      <td>4.812012</td>\n",
       "      <td>4.852539</td>\n",
       "      <td>4.972656</td>\n",
       "      <td>4.843262</td>\n",
       "      <td>...</td>\n",
       "      <td>4.786621</td>\n",
       "      <td>4.854004</td>\n",
       "      <td>4.850098</td>\n",
       "      <td>4.797363</td>\n",
       "      <td>4.797852</td>\n",
       "      <td>4.824219</td>\n",
       "      <td>4.800293</td>\n",
       "      <td>4.812012</td>\n",
       "      <td>4.872070</td>\n",
       "      <td>4.813477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.978159</td>\n",
       "      <td>11.861328</td>\n",
       "      <td>11.611328</td>\n",
       "      <td>10.238281</td>\n",
       "      <td>9.562500</td>\n",
       "      <td>10.841797</td>\n",
       "      <td>11.351562</td>\n",
       "      <td>10.564453</td>\n",
       "      <td>11.300781</td>\n",
       "      <td>9.648438</td>\n",
       "      <td>...</td>\n",
       "      <td>10.869141</td>\n",
       "      <td>10.480469</td>\n",
       "      <td>11.900391</td>\n",
       "      <td>11.300781</td>\n",
       "      <td>10.591797</td>\n",
       "      <td>11.982422</td>\n",
       "      <td>11.115234</td>\n",
       "      <td>11.400391</td>\n",
       "      <td>10.392578</td>\n",
       "      <td>10.582031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>4.740273</td>\n",
       "      <td>5.305664</td>\n",
       "      <td>4.681152</td>\n",
       "      <td>4.795410</td>\n",
       "      <td>4.846680</td>\n",
       "      <td>4.712891</td>\n",
       "      <td>4.683105</td>\n",
       "      <td>4.719238</td>\n",
       "      <td>4.760254</td>\n",
       "      <td>4.727051</td>\n",
       "      <td>...</td>\n",
       "      <td>4.730957</td>\n",
       "      <td>4.722656</td>\n",
       "      <td>4.735352</td>\n",
       "      <td>4.744141</td>\n",
       "      <td>4.723633</td>\n",
       "      <td>4.699707</td>\n",
       "      <td>4.740234</td>\n",
       "      <td>4.731445</td>\n",
       "      <td>4.727539</td>\n",
       "      <td>4.713379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>4.875744</td>\n",
       "      <td>4.892578</td>\n",
       "      <td>4.951172</td>\n",
       "      <td>4.938477</td>\n",
       "      <td>4.877930</td>\n",
       "      <td>5.062500</td>\n",
       "      <td>4.871094</td>\n",
       "      <td>4.910156</td>\n",
       "      <td>4.992188</td>\n",
       "      <td>4.981445</td>\n",
       "      <td>...</td>\n",
       "      <td>4.886719</td>\n",
       "      <td>4.900391</td>\n",
       "      <td>4.869141</td>\n",
       "      <td>4.910156</td>\n",
       "      <td>4.896484</td>\n",
       "      <td>4.831055</td>\n",
       "      <td>4.923828</td>\n",
       "      <td>5.066406</td>\n",
       "      <td>5.002930</td>\n",
       "      <td>4.890625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>4.680315</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>4.670898</td>\n",
       "      <td>4.657715</td>\n",
       "      <td>4.712891</td>\n",
       "      <td>4.703125</td>\n",
       "      <td>4.690918</td>\n",
       "      <td>4.664551</td>\n",
       "      <td>4.665039</td>\n",
       "      <td>4.683594</td>\n",
       "      <td>...</td>\n",
       "      <td>4.670410</td>\n",
       "      <td>4.671387</td>\n",
       "      <td>4.714355</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>4.674316</td>\n",
       "      <td>4.694336</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>4.666992</td>\n",
       "      <td>4.657227</td>\n",
       "      <td>4.693359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>6.612026</td>\n",
       "      <td>6.639160</td>\n",
       "      <td>6.410156</td>\n",
       "      <td>6.420166</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>6.626953</td>\n",
       "      <td>6.472412</td>\n",
       "      <td>6.822266</td>\n",
       "      <td>6.800537</td>\n",
       "      <td>6.394043</td>\n",
       "      <td>...</td>\n",
       "      <td>6.310547</td>\n",
       "      <td>6.310059</td>\n",
       "      <td>6.320801</td>\n",
       "      <td>6.378174</td>\n",
       "      <td>6.380615</td>\n",
       "      <td>6.425537</td>\n",
       "      <td>6.324707</td>\n",
       "      <td>6.276855</td>\n",
       "      <td>6.554199</td>\n",
       "      <td>6.743652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>6.074064</td>\n",
       "      <td>7.055664</td>\n",
       "      <td>6.838867</td>\n",
       "      <td>6.343750</td>\n",
       "      <td>6.431152</td>\n",
       "      <td>6.651855</td>\n",
       "      <td>6.940430</td>\n",
       "      <td>6.520508</td>\n",
       "      <td>7.701660</td>\n",
       "      <td>6.390137</td>\n",
       "      <td>...</td>\n",
       "      <td>6.027832</td>\n",
       "      <td>6.108398</td>\n",
       "      <td>7.544434</td>\n",
       "      <td>6.250977</td>\n",
       "      <td>6.095703</td>\n",
       "      <td>6.333496</td>\n",
       "      <td>6.608398</td>\n",
       "      <td>6.477051</td>\n",
       "      <td>6.934570</td>\n",
       "      <td>6.090332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>5.904870</td>\n",
       "      <td>5.963867</td>\n",
       "      <td>6.434082</td>\n",
       "      <td>5.895508</td>\n",
       "      <td>5.866699</td>\n",
       "      <td>6.193359</td>\n",
       "      <td>5.788086</td>\n",
       "      <td>6.060547</td>\n",
       "      <td>5.915039</td>\n",
       "      <td>6.078613</td>\n",
       "      <td>...</td>\n",
       "      <td>6.214355</td>\n",
       "      <td>6.235352</td>\n",
       "      <td>6.056152</td>\n",
       "      <td>6.332520</td>\n",
       "      <td>5.993164</td>\n",
       "      <td>6.766602</td>\n",
       "      <td>5.933594</td>\n",
       "      <td>6.520996</td>\n",
       "      <td>6.223633</td>\n",
       "      <td>5.875000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>7.978461</td>\n",
       "      <td>7.258789</td>\n",
       "      <td>7.264160</td>\n",
       "      <td>7.532715</td>\n",
       "      <td>7.452637</td>\n",
       "      <td>7.616211</td>\n",
       "      <td>7.958984</td>\n",
       "      <td>8.042480</td>\n",
       "      <td>6.781250</td>\n",
       "      <td>8.026855</td>\n",
       "      <td>...</td>\n",
       "      <td>7.327148</td>\n",
       "      <td>7.410156</td>\n",
       "      <td>6.974121</td>\n",
       "      <td>7.502930</td>\n",
       "      <td>7.506348</td>\n",
       "      <td>7.829102</td>\n",
       "      <td>6.866699</td>\n",
       "      <td>6.866211</td>\n",
       "      <td>7.709961</td>\n",
       "      <td>7.666992</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17667 rows Ã— 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM2411058  GSM2411059  GSM2411060  GSM2411061  GSM2411062  \\\n",
       "A1BG       5.106244    5.238281    5.280273    5.072266    5.064453   \n",
       "A1CF       4.872861    4.875000    4.913086    4.846191    4.951660   \n",
       "A2M        9.978159   11.861328   11.611328   10.238281    9.562500   \n",
       "A2ML1      4.740273    5.305664    4.681152    4.795410    4.846680   \n",
       "A3GALT2    4.875744    4.892578    4.951172    4.938477    4.877930   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11A     4.680315    4.662598    4.670898    4.657715    4.712891   \n",
       "ZYG11B     6.612026    6.639160    6.410156    6.420166    6.500000   \n",
       "ZYX        6.074064    7.055664    6.838867    6.343750    6.431152   \n",
       "ZZEF1      5.904870    5.963867    6.434082    5.895508    5.866699   \n",
       "ZZZ3       7.978461    7.258789    7.264160    7.532715    7.452637   \n",
       "\n",
       "         GSM2411063  GSM2411064  GSM2411065  GSM2411066  GSM2411067  ...  \\\n",
       "A1BG       5.056641    6.063477    4.973633    5.327148    5.094727  ...   \n",
       "A1CF       4.916016    4.812012    4.852539    4.972656    4.843262  ...   \n",
       "A2M       10.841797   11.351562   10.564453   11.300781    9.648438  ...   \n",
       "A2ML1      4.712891    4.683105    4.719238    4.760254    4.727051  ...   \n",
       "A3GALT2    5.062500    4.871094    4.910156    4.992188    4.981445  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A     4.703125    4.690918    4.664551    4.665039    4.683594  ...   \n",
       "ZYG11B     6.626953    6.472412    6.822266    6.800537    6.394043  ...   \n",
       "ZYX        6.651855    6.940430    6.520508    7.701660    6.390137  ...   \n",
       "ZZEF1      6.193359    5.788086    6.060547    5.915039    6.078613  ...   \n",
       "ZZZ3       7.616211    7.958984    8.042480    6.781250    8.026855  ...   \n",
       "\n",
       "         GSM2411111  GSM2411112  GSM2411113  GSM2411114  GSM2411115  \\\n",
       "A1BG       5.244141    4.891602    5.267578    5.543945    5.151367   \n",
       "A1CF       4.786621    4.854004    4.850098    4.797363    4.797852   \n",
       "A2M       10.869141   10.480469   11.900391   11.300781   10.591797   \n",
       "A2ML1      4.730957    4.722656    4.735352    4.744141    4.723633   \n",
       "A3GALT2    4.886719    4.900391    4.869141    4.910156    4.896484   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11A     4.670410    4.671387    4.714355    4.662598    4.674316   \n",
       "ZYG11B     6.310547    6.310059    6.320801    6.378174    6.380615   \n",
       "ZYX        6.027832    6.108398    7.544434    6.250977    6.095703   \n",
       "ZZEF1      6.214355    6.235352    6.056152    6.332520    5.993164   \n",
       "ZZZ3       7.327148    7.410156    6.974121    7.502930    7.506348   \n",
       "\n",
       "         GSM2411116  GSM2411117  GSM2411118  GSM2411119  GSM2411120  \n",
       "A1BG       5.382812    5.319336    4.903320    5.154297    6.026367  \n",
       "A1CF       4.824219    4.800293    4.812012    4.872070    4.813477  \n",
       "A2M       11.982422   11.115234   11.400391   10.392578   10.582031  \n",
       "A2ML1      4.699707    4.740234    4.731445    4.727539    4.713379  \n",
       "A3GALT2    4.831055    4.923828    5.066406    5.002930    4.890625  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZYG11A     4.694336    4.662598    4.666992    4.657227    4.693359  \n",
       "ZYG11B     6.425537    6.324707    6.276855    6.554199    6.743652  \n",
       "ZYX        6.333496    6.608398    6.477051    6.934570    6.090332  \n",
       "ZZEF1      6.766602    5.933594    6.520996    6.223633    5.875000  \n",
       "ZZZ3       7.829102    6.866699    6.866211    7.709961    7.666992  \n",
       "\n",
       "[17667 rows x 63 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'condition: tumor':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2411058</th>\n",
       "      <th>GSM2411059</th>\n",
       "      <th>GSM2411060</th>\n",
       "      <th>GSM2411061</th>\n",
       "      <th>GSM2411062</th>\n",
       "      <th>GSM2411063</th>\n",
       "      <th>GSM2411064</th>\n",
       "      <th>GSM2411065</th>\n",
       "      <th>GSM2411066</th>\n",
       "      <th>GSM2411067</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2411111</th>\n",
       "      <th>GSM2411112</th>\n",
       "      <th>GSM2411113</th>\n",
       "      <th>GSM2411114</th>\n",
       "      <th>GSM2411115</th>\n",
       "      <th>GSM2411116</th>\n",
       "      <th>GSM2411117</th>\n",
       "      <th>GSM2411118</th>\n",
       "      <th>GSM2411119</th>\n",
       "      <th>GSM2411120</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM2411058  GSM2411059  GSM2411060  GSM2411061  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM2411062  GSM2411063  GSM2411064  GSM2411065  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM2411066  GSM2411067  ...  GSM2411111  GSM2411112  \\\n",
       "Adrenocortical Cancer           1           1  ...           1           1   \n",
       "\n",
       "                       GSM2411113  GSM2411114  GSM2411115  GSM2411116  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM2411117  GSM2411118  GSM2411119  GSM2411120  \n",
       "Adrenocortical Cancer           1           1           1           1  \n",
       "\n",
       "[1 rows x 63 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2411058</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.106244</td>\n",
       "      <td>4.872861</td>\n",
       "      <td>9.978159</td>\n",
       "      <td>4.740273</td>\n",
       "      <td>4.875744</td>\n",
       "      <td>5.195803</td>\n",
       "      <td>4.952567</td>\n",
       "      <td>6.198320</td>\n",
       "      <td>6.548022</td>\n",
       "      <td>...</td>\n",
       "      <td>7.962596</td>\n",
       "      <td>6.626484</td>\n",
       "      <td>9.489847</td>\n",
       "      <td>5.142927</td>\n",
       "      <td>5.985722</td>\n",
       "      <td>4.680315</td>\n",
       "      <td>6.612026</td>\n",
       "      <td>6.074064</td>\n",
       "      <td>5.904870</td>\n",
       "      <td>7.978461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411059</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.238281</td>\n",
       "      <td>4.875000</td>\n",
       "      <td>11.861328</td>\n",
       "      <td>5.305664</td>\n",
       "      <td>4.892578</td>\n",
       "      <td>5.055664</td>\n",
       "      <td>4.983398</td>\n",
       "      <td>6.668945</td>\n",
       "      <td>5.909424</td>\n",
       "      <td>...</td>\n",
       "      <td>7.197266</td>\n",
       "      <td>5.800293</td>\n",
       "      <td>7.072754</td>\n",
       "      <td>5.172852</td>\n",
       "      <td>5.607910</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>6.639160</td>\n",
       "      <td>7.055664</td>\n",
       "      <td>5.963867</td>\n",
       "      <td>7.258789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411060</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.280273</td>\n",
       "      <td>4.913086</td>\n",
       "      <td>11.611328</td>\n",
       "      <td>4.681152</td>\n",
       "      <td>4.951172</td>\n",
       "      <td>5.188802</td>\n",
       "      <td>4.914062</td>\n",
       "      <td>7.054688</td>\n",
       "      <td>6.372559</td>\n",
       "      <td>...</td>\n",
       "      <td>6.713867</td>\n",
       "      <td>5.827881</td>\n",
       "      <td>7.602051</td>\n",
       "      <td>5.137695</td>\n",
       "      <td>6.056348</td>\n",
       "      <td>4.670898</td>\n",
       "      <td>6.410156</td>\n",
       "      <td>6.838867</td>\n",
       "      <td>6.434082</td>\n",
       "      <td>7.264160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411061</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.072266</td>\n",
       "      <td>4.846191</td>\n",
       "      <td>10.238281</td>\n",
       "      <td>4.795410</td>\n",
       "      <td>4.938477</td>\n",
       "      <td>5.052409</td>\n",
       "      <td>5.004883</td>\n",
       "      <td>6.350586</td>\n",
       "      <td>6.262939</td>\n",
       "      <td>...</td>\n",
       "      <td>7.292969</td>\n",
       "      <td>5.847900</td>\n",
       "      <td>7.145996</td>\n",
       "      <td>5.144043</td>\n",
       "      <td>5.725586</td>\n",
       "      <td>4.657715</td>\n",
       "      <td>6.420166</td>\n",
       "      <td>6.343750</td>\n",
       "      <td>5.895508</td>\n",
       "      <td>7.532715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411062</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.064453</td>\n",
       "      <td>4.951660</td>\n",
       "      <td>9.562500</td>\n",
       "      <td>4.846680</td>\n",
       "      <td>4.877930</td>\n",
       "      <td>5.063151</td>\n",
       "      <td>4.928711</td>\n",
       "      <td>6.232422</td>\n",
       "      <td>6.250488</td>\n",
       "      <td>...</td>\n",
       "      <td>7.392578</td>\n",
       "      <td>5.827148</td>\n",
       "      <td>7.178223</td>\n",
       "      <td>5.222168</td>\n",
       "      <td>5.692285</td>\n",
       "      <td>4.712891</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>6.431152</td>\n",
       "      <td>5.866699</td>\n",
       "      <td>7.452637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411116</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.382812</td>\n",
       "      <td>4.824219</td>\n",
       "      <td>11.982422</td>\n",
       "      <td>4.699707</td>\n",
       "      <td>4.831055</td>\n",
       "      <td>5.071615</td>\n",
       "      <td>4.986328</td>\n",
       "      <td>6.214844</td>\n",
       "      <td>6.644043</td>\n",
       "      <td>...</td>\n",
       "      <td>7.006836</td>\n",
       "      <td>6.724854</td>\n",
       "      <td>7.033691</td>\n",
       "      <td>5.236816</td>\n",
       "      <td>6.435742</td>\n",
       "      <td>4.694336</td>\n",
       "      <td>6.425537</td>\n",
       "      <td>6.333496</td>\n",
       "      <td>6.766602</td>\n",
       "      <td>7.829102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411117</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.319336</td>\n",
       "      <td>4.800293</td>\n",
       "      <td>11.115234</td>\n",
       "      <td>4.740234</td>\n",
       "      <td>4.923828</td>\n",
       "      <td>5.130534</td>\n",
       "      <td>4.795898</td>\n",
       "      <td>6.176758</td>\n",
       "      <td>6.421875</td>\n",
       "      <td>...</td>\n",
       "      <td>6.976562</td>\n",
       "      <td>5.992188</td>\n",
       "      <td>7.158203</td>\n",
       "      <td>5.168945</td>\n",
       "      <td>5.774121</td>\n",
       "      <td>4.662598</td>\n",
       "      <td>6.324707</td>\n",
       "      <td>6.608398</td>\n",
       "      <td>5.933594</td>\n",
       "      <td>6.866699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411118</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.903320</td>\n",
       "      <td>4.812012</td>\n",
       "      <td>11.400391</td>\n",
       "      <td>4.731445</td>\n",
       "      <td>5.066406</td>\n",
       "      <td>5.210286</td>\n",
       "      <td>4.906250</td>\n",
       "      <td>6.487305</td>\n",
       "      <td>5.935059</td>\n",
       "      <td>...</td>\n",
       "      <td>6.934570</td>\n",
       "      <td>7.105957</td>\n",
       "      <td>8.438965</td>\n",
       "      <td>5.033203</td>\n",
       "      <td>5.518555</td>\n",
       "      <td>4.666992</td>\n",
       "      <td>6.276855</td>\n",
       "      <td>6.477051</td>\n",
       "      <td>6.520996</td>\n",
       "      <td>6.866211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411119</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.154297</td>\n",
       "      <td>4.872070</td>\n",
       "      <td>10.392578</td>\n",
       "      <td>4.727539</td>\n",
       "      <td>5.002930</td>\n",
       "      <td>5.390625</td>\n",
       "      <td>4.973633</td>\n",
       "      <td>6.874023</td>\n",
       "      <td>6.340088</td>\n",
       "      <td>...</td>\n",
       "      <td>6.652344</td>\n",
       "      <td>5.964600</td>\n",
       "      <td>7.772949</td>\n",
       "      <td>5.078613</td>\n",
       "      <td>5.781250</td>\n",
       "      <td>4.657227</td>\n",
       "      <td>6.554199</td>\n",
       "      <td>6.934570</td>\n",
       "      <td>6.223633</td>\n",
       "      <td>7.709961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2411120</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.026367</td>\n",
       "      <td>4.813477</td>\n",
       "      <td>10.582031</td>\n",
       "      <td>4.713379</td>\n",
       "      <td>4.890625</td>\n",
       "      <td>5.053711</td>\n",
       "      <td>4.974609</td>\n",
       "      <td>6.042969</td>\n",
       "      <td>6.013916</td>\n",
       "      <td>...</td>\n",
       "      <td>7.151367</td>\n",
       "      <td>5.941895</td>\n",
       "      <td>7.477051</td>\n",
       "      <td>5.024414</td>\n",
       "      <td>5.707422</td>\n",
       "      <td>4.693359</td>\n",
       "      <td>6.743652</td>\n",
       "      <td>6.090332</td>\n",
       "      <td>5.875000</td>\n",
       "      <td>7.666992</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 17668 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Adrenocortical Cancer      A1BG      A1CF        A2M     A2ML1  \\\n",
       "GSM2411058                    1.0  5.106244  4.872861   9.978159  4.740273   \n",
       "GSM2411059                    1.0  5.238281  4.875000  11.861328  5.305664   \n",
       "GSM2411060                    1.0  5.280273  4.913086  11.611328  4.681152   \n",
       "GSM2411061                    1.0  5.072266  4.846191  10.238281  4.795410   \n",
       "GSM2411062                    1.0  5.064453  4.951660   9.562500  4.846680   \n",
       "...                           ...       ...       ...        ...       ...   \n",
       "GSM2411116                    1.0  5.382812  4.824219  11.982422  4.699707   \n",
       "GSM2411117                    1.0  5.319336  4.800293  11.115234  4.740234   \n",
       "GSM2411118                    1.0  4.903320  4.812012  11.400391  4.731445   \n",
       "GSM2411119                    1.0  5.154297  4.872070  10.392578  4.727539   \n",
       "GSM2411120                    1.0  6.026367  4.813477  10.582031  4.713379   \n",
       "\n",
       "             A3GALT2    A4GALT     A4GNT      AAAS      AACS  ...      ZW10  \\\n",
       "GSM2411058  4.875744  5.195803  4.952567  6.198320  6.548022  ...  7.962596   \n",
       "GSM2411059  4.892578  5.055664  4.983398  6.668945  5.909424  ...  7.197266   \n",
       "GSM2411060  4.951172  5.188802  4.914062  7.054688  6.372559  ...  6.713867   \n",
       "GSM2411061  4.938477  5.052409  5.004883  6.350586  6.262939  ...  7.292969   \n",
       "GSM2411062  4.877930  5.063151  4.928711  6.232422  6.250488  ...  7.392578   \n",
       "...              ...       ...       ...       ...       ...  ...       ...   \n",
       "GSM2411116  4.831055  5.071615  4.986328  6.214844  6.644043  ...  7.006836   \n",
       "GSM2411117  4.923828  5.130534  4.795898  6.176758  6.421875  ...  6.976562   \n",
       "GSM2411118  5.066406  5.210286  4.906250  6.487305  5.935059  ...  6.934570   \n",
       "GSM2411119  5.002930  5.390625  4.973633  6.874023  6.340088  ...  6.652344   \n",
       "GSM2411120  4.890625  5.053711  4.974609  6.042969  6.013916  ...  7.151367   \n",
       "\n",
       "              ZWILCH     ZWINT      ZXDB      ZXDC    ZYG11A    ZYG11B  \\\n",
       "GSM2411058  6.626484  9.489847  5.142927  5.985722  4.680315  6.612026   \n",
       "GSM2411059  5.800293  7.072754  5.172852  5.607910  4.662598  6.639160   \n",
       "GSM2411060  5.827881  7.602051  5.137695  6.056348  4.670898  6.410156   \n",
       "GSM2411061  5.847900  7.145996  5.144043  5.725586  4.657715  6.420166   \n",
       "GSM2411062  5.827148  7.178223  5.222168  5.692285  4.712891  6.500000   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "GSM2411116  6.724854  7.033691  5.236816  6.435742  4.694336  6.425537   \n",
       "GSM2411117  5.992188  7.158203  5.168945  5.774121  4.662598  6.324707   \n",
       "GSM2411118  7.105957  8.438965  5.033203  5.518555  4.666992  6.276855   \n",
       "GSM2411119  5.964600  7.772949  5.078613  5.781250  4.657227  6.554199   \n",
       "GSM2411120  5.941895  7.477051  5.024414  5.707422  4.693359  6.743652   \n",
       "\n",
       "                 ZYX     ZZEF1      ZZZ3  \n",
       "GSM2411058  6.074064  5.904870  7.978461  \n",
       "GSM2411059  7.055664  5.963867  7.258789  \n",
       "GSM2411060  6.838867  6.434082  7.264160  \n",
       "GSM2411061  6.343750  5.895508  7.532715  \n",
       "GSM2411062  6.431152  5.866699  7.452637  \n",
       "...              ...       ...       ...  \n",
       "GSM2411116  6.333496  6.766602  7.829102  \n",
       "GSM2411117  6.608398  5.933594  6.866699  \n",
       "GSM2411118  6.477051  6.520996  6.866211  \n",
       "GSM2411119  6.934570  6.223633  7.709961  \n",
       "GSM2411120  6.090332  5.875000  7.666992  \n",
       "\n",
       "[63 rows x 17668 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 63 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '0.0' with 5 occurrences. This represents 7.94% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Genome wide DNA Methylation analysis of benign and malignant adrenocortical tumors\"\n",
      "!Series_summary\t\"Genome wide DNA methylation profiling of normal adrenocortical tissue, adrenocortical adenomas and adrenocortical carcinomas. The Illumina Infinium 27k Human DNA methylation Beadchip v1.2 was used to obtain DNA methylation profiles. Samples included 6 normal adrenocortical tissue samples, 27 adenomas and 15 carcinomas.\"\n",
      "!Series_overall_design\t\"Bisulphite converted DNA from the 48 samples were hybridised to the Illumina Infinium 27k Human Methylation Beadchip v1.2\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM889444</th>\n",
       "      <th>GSM889445</th>\n",
       "      <th>GSM889446</th>\n",
       "      <th>GSM889447</th>\n",
       "      <th>GSM889448</th>\n",
       "      <th>GSM889449</th>\n",
       "      <th>GSM889450</th>\n",
       "      <th>GSM889451</th>\n",
       "      <th>GSM889452</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM889482</th>\n",
       "      <th>GSM889483</th>\n",
       "      <th>GSM889484</th>\n",
       "      <th>GSM889485</th>\n",
       "      <th>GSM889486</th>\n",
       "      <th>GSM889487</th>\n",
       "      <th>GSM889488</th>\n",
       "      <th>GSM889489</th>\n",
       "      <th>GSM889490</th>\n",
       "      <th>GSM889491</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "      <td>tissue: adrenal tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: adenoma</td>\n",
       "      <td>disease state: adenoma</td>\n",
       "      <td>disease state: adenoma</td>\n",
       "      <td>disease state: adenoma</td>\n",
       "      <td>disease state: normal adrenal tissue</td>\n",
       "      <td>disease state: normal adrenal tissue</td>\n",
       "      <td>disease state: normal adrenal tissue</td>\n",
       "      <td>disease state: normal adrenal tissue</td>\n",
       "      <td>disease state: normal adrenal tissue</td>\n",
       "      <td>disease state: normal adrenal tissue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                 GSM889444  \\\n",
       "0  !Sample_characteristics_ch1            gender: Female   \n",
       "1  !Sample_characteristics_ch1    tissue: adrenal tissue   \n",
       "2  !Sample_characteristics_ch1  disease state: carcinoma   \n",
       "\n",
       "                  GSM889445                 GSM889446  \\\n",
       "0            gender: Female            gender: Female   \n",
       "1    tissue: adrenal tissue    tissue: adrenal tissue   \n",
       "2  disease state: carcinoma  disease state: carcinoma   \n",
       "\n",
       "                  GSM889447                 GSM889448  \\\n",
       "0            gender: Female            gender: Female   \n",
       "1    tissue: adrenal tissue    tissue: adrenal tissue   \n",
       "2  disease state: carcinoma  disease state: carcinoma   \n",
       "\n",
       "                  GSM889449                 GSM889450  \\\n",
       "0              gender: Male            gender: Female   \n",
       "1    tissue: adrenal tissue    tissue: adrenal tissue   \n",
       "2  disease state: carcinoma  disease state: carcinoma   \n",
       "\n",
       "                  GSM889451                 GSM889452  ...  \\\n",
       "0              gender: Male              gender: Male  ...   \n",
       "1    tissue: adrenal tissue    tissue: adrenal tissue  ...   \n",
       "2  disease state: carcinoma  disease state: carcinoma  ...   \n",
       "\n",
       "                GSM889482               GSM889483               GSM889484  \\\n",
       "0          gender: Female          gender: Female          gender: Female   \n",
       "1  tissue: adrenal tissue  tissue: adrenal tissue  tissue: adrenal tissue   \n",
       "2  disease state: adenoma  disease state: adenoma  disease state: adenoma   \n",
       "\n",
       "                GSM889485                             GSM889486  \\\n",
       "0            gender: Male                          gender: Male   \n",
       "1  tissue: adrenal tissue                tissue: adrenal tissue   \n",
       "2  disease state: adenoma  disease state: normal adrenal tissue   \n",
       "\n",
       "                              GSM889487                             GSM889488  \\\n",
       "0                        gender: Female                          gender: Male   \n",
       "1                tissue: adrenal tissue                tissue: adrenal tissue   \n",
       "2  disease state: normal adrenal tissue  disease state: normal adrenal tissue   \n",
       "\n",
       "                              GSM889489                             GSM889490  \\\n",
       "0                        gender: Female                          gender: Male   \n",
       "1                tissue: adrenal tissue                tissue: adrenal tissue   \n",
       "2  disease state: normal adrenal tissue  disease state: normal adrenal tissue   \n",
       "\n",
       "                              GSM889491  \n",
       "0                        gender: Female  \n",
       "1                tissue: adrenal tissue  \n",
       "2  disease state: normal adrenal tissue  \n",
       "\n",
       "[3 rows x 49 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE36353\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease state: carcinoma',\n",
       "       'disease state: adenoma', 'disease state: normal adrenal tissue'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM889444</th>\n",
       "      <th>GSM889445</th>\n",
       "      <th>GSM889446</th>\n",
       "      <th>GSM889447</th>\n",
       "      <th>GSM889448</th>\n",
       "      <th>GSM889449</th>\n",
       "      <th>GSM889450</th>\n",
       "      <th>GSM889451</th>\n",
       "      <th>GSM889452</th>\n",
       "      <th>GSM889453</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM889482</th>\n",
       "      <th>GSM889483</th>\n",
       "      <th>GSM889484</th>\n",
       "      <th>GSM889485</th>\n",
       "      <th>GSM889486</th>\n",
       "      <th>GSM889487</th>\n",
       "      <th>GSM889488</th>\n",
       "      <th>GSM889489</th>\n",
       "      <th>GSM889490</th>\n",
       "      <th>GSM889491</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000292</th>\n",
       "      <td>0.590574</td>\n",
       "      <td>0.513603</td>\n",
       "      <td>0.826171</td>\n",
       "      <td>0.867341</td>\n",
       "      <td>0.598491</td>\n",
       "      <td>0.785434</td>\n",
       "      <td>0.566842</td>\n",
       "      <td>0.801596</td>\n",
       "      <td>0.782646</td>\n",
       "      <td>0.713891</td>\n",
       "      <td>...</td>\n",
       "      <td>0.573703</td>\n",
       "      <td>0.464043</td>\n",
       "      <td>0.399872</td>\n",
       "      <td>0.648043</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>0.601324</td>\n",
       "      <td>0.522049</td>\n",
       "      <td>0.534815</td>\n",
       "      <td>0.586788</td>\n",
       "      <td>0.606669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00002426</th>\n",
       "      <td>0.256898</td>\n",
       "      <td>0.105301</td>\n",
       "      <td>0.120422</td>\n",
       "      <td>0.047387</td>\n",
       "      <td>0.158108</td>\n",
       "      <td>0.103077</td>\n",
       "      <td>0.553597</td>\n",
       "      <td>0.270270</td>\n",
       "      <td>0.117484</td>\n",
       "      <td>0.130336</td>\n",
       "      <td>...</td>\n",
       "      <td>0.271988</td>\n",
       "      <td>0.279047</td>\n",
       "      <td>0.216480</td>\n",
       "      <td>0.335722</td>\n",
       "      <td>0.298562</td>\n",
       "      <td>0.293625</td>\n",
       "      <td>0.334871</td>\n",
       "      <td>0.321670</td>\n",
       "      <td>0.222039</td>\n",
       "      <td>0.293077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00003994</th>\n",
       "      <td>0.078937</td>\n",
       "      <td>0.095357</td>\n",
       "      <td>0.066558</td>\n",
       "      <td>0.148379</td>\n",
       "      <td>0.302606</td>\n",
       "      <td>0.073422</td>\n",
       "      <td>0.082058</td>\n",
       "      <td>0.103124</td>\n",
       "      <td>0.062960</td>\n",
       "      <td>0.303335</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132663</td>\n",
       "      <td>0.222629</td>\n",
       "      <td>0.102947</td>\n",
       "      <td>0.080621</td>\n",
       "      <td>0.095720</td>\n",
       "      <td>0.093675</td>\n",
       "      <td>0.089851</td>\n",
       "      <td>0.082194</td>\n",
       "      <td>0.095572</td>\n",
       "      <td>0.083108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00005847</th>\n",
       "      <td>0.684397</td>\n",
       "      <td>0.808231</td>\n",
       "      <td>0.740177</td>\n",
       "      <td>0.798983</td>\n",
       "      <td>0.600528</td>\n",
       "      <td>0.769102</td>\n",
       "      <td>0.388342</td>\n",
       "      <td>0.522643</td>\n",
       "      <td>0.751418</td>\n",
       "      <td>0.213138</td>\n",
       "      <td>...</td>\n",
       "      <td>0.496777</td>\n",
       "      <td>0.231583</td>\n",
       "      <td>0.250337</td>\n",
       "      <td>0.429693</td>\n",
       "      <td>0.630685</td>\n",
       "      <td>0.597885</td>\n",
       "      <td>0.663874</td>\n",
       "      <td>0.678045</td>\n",
       "      <td>0.581554</td>\n",
       "      <td>0.648205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00006414</th>\n",
       "      <td>0.113754</td>\n",
       "      <td>0.094297</td>\n",
       "      <td>0.068810</td>\n",
       "      <td>0.095124</td>\n",
       "      <td>0.099132</td>\n",
       "      <td>0.045772</td>\n",
       "      <td>0.130852</td>\n",
       "      <td>0.078138</td>\n",
       "      <td>0.090765</td>\n",
       "      <td>0.092367</td>\n",
       "      <td>...</td>\n",
       "      <td>0.157187</td>\n",
       "      <td>0.103861</td>\n",
       "      <td>0.148690</td>\n",
       "      <td>0.086650</td>\n",
       "      <td>0.102798</td>\n",
       "      <td>0.095674</td>\n",
       "      <td>0.085801</td>\n",
       "      <td>0.095066</td>\n",
       "      <td>0.097189</td>\n",
       "      <td>0.121290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27657249</th>\n",
       "      <td>0.167368</td>\n",
       "      <td>0.163048</td>\n",
       "      <td>0.170360</td>\n",
       "      <td>0.206942</td>\n",
       "      <td>0.275417</td>\n",
       "      <td>0.182009</td>\n",
       "      <td>0.162453</td>\n",
       "      <td>0.198111</td>\n",
       "      <td>0.183755</td>\n",
       "      <td>0.193382</td>\n",
       "      <td>...</td>\n",
       "      <td>0.209191</td>\n",
       "      <td>0.212529</td>\n",
       "      <td>0.200447</td>\n",
       "      <td>0.225882</td>\n",
       "      <td>0.157167</td>\n",
       "      <td>0.159855</td>\n",
       "      <td>0.175443</td>\n",
       "      <td>0.169527</td>\n",
       "      <td>0.177136</td>\n",
       "      <td>0.149817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27657283</th>\n",
       "      <td>0.082578</td>\n",
       "      <td>0.110022</td>\n",
       "      <td>0.102603</td>\n",
       "      <td>0.111641</td>\n",
       "      <td>0.099657</td>\n",
       "      <td>0.078547</td>\n",
       "      <td>0.147287</td>\n",
       "      <td>0.096087</td>\n",
       "      <td>0.111524</td>\n",
       "      <td>0.118322</td>\n",
       "      <td>...</td>\n",
       "      <td>0.235684</td>\n",
       "      <td>0.106418</td>\n",
       "      <td>0.141741</td>\n",
       "      <td>0.097420</td>\n",
       "      <td>0.093500</td>\n",
       "      <td>0.113155</td>\n",
       "      <td>0.081989</td>\n",
       "      <td>0.092765</td>\n",
       "      <td>0.089083</td>\n",
       "      <td>0.116853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27661264</th>\n",
       "      <td>0.352964</td>\n",
       "      <td>0.230991</td>\n",
       "      <td>0.443879</td>\n",
       "      <td>0.258413</td>\n",
       "      <td>0.444090</td>\n",
       "      <td>0.490304</td>\n",
       "      <td>0.374893</td>\n",
       "      <td>0.559374</td>\n",
       "      <td>0.494604</td>\n",
       "      <td>0.399411</td>\n",
       "      <td>...</td>\n",
       "      <td>0.379100</td>\n",
       "      <td>0.509235</td>\n",
       "      <td>0.329875</td>\n",
       "      <td>0.565987</td>\n",
       "      <td>0.378737</td>\n",
       "      <td>0.298566</td>\n",
       "      <td>0.371765</td>\n",
       "      <td>0.343241</td>\n",
       "      <td>0.274428</td>\n",
       "      <td>0.362638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27662379</th>\n",
       "      <td>0.037021</td>\n",
       "      <td>0.035945</td>\n",
       "      <td>0.042062</td>\n",
       "      <td>0.025984</td>\n",
       "      <td>0.036624</td>\n",
       "      <td>0.026747</td>\n",
       "      <td>0.043812</td>\n",
       "      <td>0.030194</td>\n",
       "      <td>0.037006</td>\n",
       "      <td>0.024747</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034847</td>\n",
       "      <td>0.034346</td>\n",
       "      <td>0.049740</td>\n",
       "      <td>0.018274</td>\n",
       "      <td>0.042089</td>\n",
       "      <td>0.040415</td>\n",
       "      <td>0.033850</td>\n",
       "      <td>0.037650</td>\n",
       "      <td>0.030767</td>\n",
       "      <td>0.030493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27662877</th>\n",
       "      <td>0.061393</td>\n",
       "      <td>0.293117</td>\n",
       "      <td>0.060781</td>\n",
       "      <td>0.060578</td>\n",
       "      <td>0.070062</td>\n",
       "      <td>0.050918</td>\n",
       "      <td>0.071245</td>\n",
       "      <td>0.058902</td>\n",
       "      <td>0.076532</td>\n",
       "      <td>0.071186</td>\n",
       "      <td>...</td>\n",
       "      <td>0.103016</td>\n",
       "      <td>0.079496</td>\n",
       "      <td>0.090092</td>\n",
       "      <td>0.067224</td>\n",
       "      <td>0.055655</td>\n",
       "      <td>0.068349</td>\n",
       "      <td>0.050889</td>\n",
       "      <td>0.057360</td>\n",
       "      <td>0.050880</td>\n",
       "      <td>0.056815</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27262 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM889444  GSM889445  GSM889446  GSM889447  GSM889448  GSM889449  \\\n",
       "ID                                                                             \n",
       "cg00000292   0.590574   0.513603   0.826171   0.867341   0.598491   0.785434   \n",
       "cg00002426   0.256898   0.105301   0.120422   0.047387   0.158108   0.103077   \n",
       "cg00003994   0.078937   0.095357   0.066558   0.148379   0.302606   0.073422   \n",
       "cg00005847   0.684397   0.808231   0.740177   0.798983   0.600528   0.769102   \n",
       "cg00006414   0.113754   0.094297   0.068810   0.095124   0.099132   0.045772   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "cg27657249   0.167368   0.163048   0.170360   0.206942   0.275417   0.182009   \n",
       "cg27657283   0.082578   0.110022   0.102603   0.111641   0.099657   0.078547   \n",
       "cg27661264   0.352964   0.230991   0.443879   0.258413   0.444090   0.490304   \n",
       "cg27662379   0.037021   0.035945   0.042062   0.025984   0.036624   0.026747   \n",
       "cg27662877   0.061393   0.293117   0.060781   0.060578   0.070062   0.050918   \n",
       "\n",
       "            GSM889450  GSM889451  GSM889452  GSM889453  ...  GSM889482  \\\n",
       "ID                                                      ...              \n",
       "cg00000292   0.566842   0.801596   0.782646   0.713891  ...   0.573703   \n",
       "cg00002426   0.553597   0.270270   0.117484   0.130336  ...   0.271988   \n",
       "cg00003994   0.082058   0.103124   0.062960   0.303335  ...   0.132663   \n",
       "cg00005847   0.388342   0.522643   0.751418   0.213138  ...   0.496777   \n",
       "cg00006414   0.130852   0.078138   0.090765   0.092367  ...   0.157187   \n",
       "...               ...        ...        ...        ...  ...        ...   \n",
       "cg27657249   0.162453   0.198111   0.183755   0.193382  ...   0.209191   \n",
       "cg27657283   0.147287   0.096087   0.111524   0.118322  ...   0.235684   \n",
       "cg27661264   0.374893   0.559374   0.494604   0.399411  ...   0.379100   \n",
       "cg27662379   0.043812   0.030194   0.037006   0.024747  ...   0.034847   \n",
       "cg27662877   0.071245   0.058902   0.076532   0.071186  ...   0.103016   \n",
       "\n",
       "            GSM889483  GSM889484  GSM889485  GSM889486  GSM889487  GSM889488  \\\n",
       "ID                                                                             \n",
       "cg00000292   0.464043   0.399872   0.648043   0.736842   0.601324   0.522049   \n",
       "cg00002426   0.279047   0.216480   0.335722   0.298562   0.293625   0.334871   \n",
       "cg00003994   0.222629   0.102947   0.080621   0.095720   0.093675   0.089851   \n",
       "cg00005847   0.231583   0.250337   0.429693   0.630685   0.597885   0.663874   \n",
       "cg00006414   0.103861   0.148690   0.086650   0.102798   0.095674   0.085801   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "cg27657249   0.212529   0.200447   0.225882   0.157167   0.159855   0.175443   \n",
       "cg27657283   0.106418   0.141741   0.097420   0.093500   0.113155   0.081989   \n",
       "cg27661264   0.509235   0.329875   0.565987   0.378737   0.298566   0.371765   \n",
       "cg27662379   0.034346   0.049740   0.018274   0.042089   0.040415   0.033850   \n",
       "cg27662877   0.079496   0.090092   0.067224   0.055655   0.068349   0.050889   \n",
       "\n",
       "            GSM889489  GSM889490  GSM889491  \n",
       "ID                                           \n",
       "cg00000292   0.534815   0.586788   0.606669  \n",
       "cg00002426   0.321670   0.222039   0.293077  \n",
       "cg00003994   0.082194   0.095572   0.083108  \n",
       "cg00005847   0.678045   0.581554   0.648205  \n",
       "cg00006414   0.095066   0.097189   0.121290  \n",
       "...               ...        ...        ...  \n",
       "cg27657249   0.169527   0.177136   0.149817  \n",
       "cg27657283   0.092765   0.089083   0.116853  \n",
       "cg27661264   0.343241   0.274428   0.362638  \n",
       "cg27662379   0.037650   0.030767   0.030493  \n",
       "cg27662877   0.057360   0.050880   0.056815  \n",
       "\n",
       "[27262 rows x 48 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'Name': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'IlmnStrand': ['TOP', 'TOP', 'TOP', 'BOT', 'BOT'], 'AddressA_ID': [990370.0, 6580397.0, 7150184.0, 4850717.0, 6980731.0], 'AlleleA_ProbeSeq': ['AAACATTAATTACCAACCACTCTTCCAAAAAACACTTACCATTAAAACCA', 'AATATAATAACATTACCTTACCCATCTTATAATCAAACCAAACAAAAACA', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCA', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACA', 'CTCAAAAACCAAACAAAACAAAACCCCAATACTAATCATTAATAAAATCA'], 'AddressB_ID': [6660678.0, 6100343.0, 7150392.0, 1260113.0, 4280093.0], 'AlleleB_ProbeSeq': ['AAACATTAATTACCAACCGCTCTTCCAAAAAACACTTACCATTAAAACCG', 'AATATAATAACATTACCTTACCCGTCTTATAATCAAACCAAACGAAAACG', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCG', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACG', 'CTCGAAAACCGAACAAAACAAAACCCCAATACTAATCGTTAATAAAATCG'], 'GenomeBuild': [36.0, 36.0, 36.0, 36.0, 36.0], 'Chr': ['16', '3', '7', '2', '7'], 'MapInfo': [28797601.0, 57718583.0, 15692387.0, 176737319.0, 148453770.0], 'Ploidy': ['diploid', 'diploid', 'diploid', 'diploid', 'diploid'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq'], 'SourceVersion': [36.1, 36.1, 36.1, 36.1, 36.1], 'SourceStrand': ['TOP', 'TOP', 'BOT', 'BOT', 'BOT'], 'SourceSeq': ['CGGCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCC', 'CGCTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATT', 'GGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTTCG', 'TACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAGCG', 'CTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGATCG'], 'TopGenomicSeq': ['TGGGGTGAGTGAGACCACGGGCCTCACCCCGGACCAAGTTAAGCGGAATCTGGAGAAATA[CG]GCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCCTCCTGCACCCCC', 'CCGCTGTCGACCAGCGCAGAATAATGCCACTTTTGATTGCAAAGTGCTATCAAGGAACCA[CG]CTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATTGTCCAGCGGCAT', 'GGTGGTGGTGGTGGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTT[CG]TTGGGGTATCCCGCGATTATGCAAGATGAGGAAGAAGTAGAGAGCTCGGGGTAAGACATA', 'CAGATAACTCAATACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAG[CG]CCATAAATTTGAAGGCCAATGATCGGTTTTCATGTAACGGGTGGTACTTCATACTGAAGT', 'GAACCGGCCCAGCTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGAT[CG]CGCCTGTGAATAGCCACTGCCCTCCAGCCTGGGCAACAGCCAGACCCCGTCTGTTTAATA'], 'Next_Base': ['T', 'T', 'T', 'C', 'C'], 'Color_Channel': ['Red', 'Red', 'Red', 'Grn', 'Grn'], 'TSS_Coordinate': [28797310.0, 57718214.0, 15692819.0, 176737051.0, 148454441.0], 'Gene_Strand': ['+', '+', '-', '+', '+'], 'Gene_ID': ['GeneID:487', 'GeneID:7871', 'GeneID:4223', 'GeneID:3232', 'GeneID:57541'], 'Symbol': ['ATP2A1', 'SLMAP', 'MEOX2', 'HOXD3', 'ZNF398'], 'Synonym': ['ATP2A; SERCA1;', 'SLAP; KIAA1601;', 'GAX; MOX2;', 'HOX4; HOX1D; HOX4A; Hox-4.1; MGC10470;', 'P51; P71; ZER6; KIAA1339;'], 'Accession': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'GID': ['GI:47132613', 'GI:56550042', 'GI:55956906', 'GI:23510372', 'GI:25777702'], 'Annotation': ['isoform a is encoded by transcript variant a; sarcoplasmic/endoplasmic reticulum calcium ATPase 1; calcium pump 1; SR Ca(2+)-ATPase 1; calcium-transporting ATPase sarcoplasmic reticulum type; fast twitch skeletal muscle isoform; endoplasmic reticulum class 1/2 Ca(2+) ATPase; go_component: membrane; go_component: integral to membrane; go_component: sarcoplasmic reticulum; go_component: smooth endoplasmic reticulum; go_function: ATP binding; go_function: hydrolase activity; go_function: nucleotide binding; go_function: calcium ion binding; go_function: magnesium ion binding; go_function: calcium-transporting ATPase activity; go_function: hydrolase activity; acting on acid anhydrides; catalyzing transmembrane movement of substances; go_process: metabolism; go_process: cation transport; go_process: proton transport; go_process: calcium ion transport; go_process: regulation of striated muscle contraction', 'Sarcolemmal-associated protein; go_component: integral to plasma membrane; go_component: smooth endoplasmic reticulum; go_function: unfolded protein binding; go_process: protein folding; go_process: muscle contraction', 'growth arrest-specific homeo box; go_component: nucleus; go_function: transcription factor activity; go_process: circulation; go_process: development; go_process: regulation of transcription; DNA-dependent', 'homeobox protein Hox-D3; Hox-4.1; mouse; homolog of; homeo box D3; go_component: nucleus; go_function: transcription factor activity; go_process: morphogenesis; go_process: regulation of transcription; DNA-dependent', 'isoform b is encoded by transcript variant 2; zinc finger DNA binding protein ZER6; zinc finger-estrogen receptor interaction; clone 6; zinc finger DNA binding protein p52/p71; go_component: nucleus; go_function: DNA binding; go_function: zinc ion binding; go_function: metal ion binding; go_function: transcriptional activator activity; go_process: transcription; go_process: regulation of transcription; DNA-dependent'], 'Product': ['ATPase; Ca++ transporting; fast twitch 1 isoform a', 'sarcolemma associated protein', 'mesenchyme homeo box 2', 'homeobox D3', 'zinc finger 398 isoform b'], 'Distance_to_TSS': [291.0, 369.0, 432.0, 268.0, 671.0], 'CPG_ISLAND': [True, True, True, False, True], 'CPG_ISLAND_LOCATIONS': ['16:28797486-28797825', '3:57716811-57718675', '7:15691512-15693551', nan, '7:148453584-148455804'], 'MIR_CPG_ISLAND': [nan, nan, nan, nan, nan], 'RANGE_GB': ['NC_000016.8', 'NC_000003.10', 'NC_000007.12', nan, 'NC_000007.12'], 'RANGE_START': [28797486.0, 57716811.0, 15691512.0, nan, 148453584.0], 'RANGE_END': [28797825.0, 57718675.0, 15693551.0, nan, 148455804.0], 'RANGE_STRAND': ['+', '+', '-', nan, '+'], 'GB_ACC': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'ORF': [487.0, 7871.0, 4223.0, 3232.0, 57541.0]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414', 'cg00007981', 'cg00008493', 'cg00008713', 'cg00009407', 'cg00010193', 'cg00011459', 'cg00012199', 'cg00012386', 'cg00012792', 'cg00013618', 'cg00014085', 'cg00014837', 'cg00015770', 'cg00016968', 'cg00019495']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'Name': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'IlmnStrand': ['TOP', 'TOP', 'TOP', 'BOT', 'BOT'], 'AddressA_ID': [990370.0, 6580397.0, 7150184.0, 4850717.0, 6980731.0], 'AlleleA_ProbeSeq': ['AAACATTAATTACCAACCACTCTTCCAAAAAACACTTACCATTAAAACCA', 'AATATAATAACATTACCTTACCCATCTTATAATCAAACCAAACAAAAACA', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCA', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACA', 'CTCAAAAACCAAACAAAACAAAACCCCAATACTAATCATTAATAAAATCA'], 'AddressB_ID': [6660678.0, 6100343.0, 7150392.0, 1260113.0, 4280093.0], 'AlleleB_ProbeSeq': ['AAACATTAATTACCAACCGCTCTTCCAAAAAACACTTACCATTAAAACCG', 'AATATAATAACATTACCTTACCCGTCTTATAATCAAACCAAACGAAAACG', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCG', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACG', 'CTCGAAAACCGAACAAAACAAAACCCCAATACTAATCGTTAATAAAATCG'], 'GenomeBuild': [36.0, 36.0, 36.0, 36.0, 36.0], 'Chr': ['16', '3', '7', '2', '7'], 'MapInfo': [28797601.0, 57718583.0, 15692387.0, 176737319.0, 148453770.0], 'Ploidy': ['diploid', 'diploid', 'diploid', 'diploid', 'diploid'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq'], 'SourceVersion': [36.1, 36.1, 36.1, 36.1, 36.1], 'SourceStrand': ['TOP', 'TOP', 'BOT', 'BOT', 'BOT'], 'SourceSeq': ['CGGCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCC', 'CGCTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATT', 'GGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTTCG', 'TACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAGCG', 'CTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGATCG'], 'TopGenomicSeq': ['TGGGGTGAGTGAGACCACGGGCCTCACCCCGGACCAAGTTAAGCGGAATCTGGAGAAATA[CG]GCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCCTCCTGCACCCCC', 'CCGCTGTCGACCAGCGCAGAATAATGCCACTTTTGATTGCAAAGTGCTATCAAGGAACCA[CG]CTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATTGTCCAGCGGCAT', 'GGTGGTGGTGGTGGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTT[CG]TTGGGGTATCCCGCGATTATGCAAGATGAGGAAGAAGTAGAGAGCTCGGGGTAAGACATA', 'CAGATAACTCAATACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAG[CG]CCATAAATTTGAAGGCCAATGATCGGTTTTCATGTAACGGGTGGTACTTCATACTGAAGT', 'GAACCGGCCCAGCTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGAT[CG]CGCCTGTGAATAGCCACTGCCCTCCAGCCTGGGCAACAGCCAGACCCCGTCTGTTTAATA'], 'Next_Base': ['T', 'T', 'T', 'C', 'C'], 'Color_Channel': ['Red', 'Red', 'Red', 'Grn', 'Grn'], 'TSS_Coordinate': [28797310.0, 57718214.0, 15692819.0, 176737051.0, 148454441.0], 'Gene_Strand': ['+', '+', '-', '+', '+'], 'Gene_ID': ['GeneID:487', 'GeneID:7871', 'GeneID:4223', 'GeneID:3232', 'GeneID:57541'], 'Symbol': ['ATP2A1', 'SLMAP', 'MEOX2', 'HOXD3', 'ZNF398'], 'Synonym': ['ATP2A; SERCA1;', 'SLAP; KIAA1601;', 'GAX; MOX2;', 'HOX4; HOX1D; HOX4A; Hox-4.1; MGC10470;', 'P51; P71; ZER6; KIAA1339;'], 'Accession': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'GID': ['GI:47132613', 'GI:56550042', 'GI:55956906', 'GI:23510372', 'GI:25777702'], 'Annotation': ['isoform a is encoded by transcript variant a; sarcoplasmic/endoplasmic reticulum calcium ATPase 1; calcium pump 1; SR Ca(2+)-ATPase 1; calcium-transporting ATPase sarcoplasmic reticulum type; fast twitch skeletal muscle isoform; endoplasmic reticulum class 1/2 Ca(2+) ATPase; go_component: membrane; go_component: integral to membrane; go_component: sarcoplasmic reticulum; go_component: smooth endoplasmic reticulum; go_function: ATP binding; go_function: hydrolase activity; go_function: nucleotide binding; go_function: calcium ion binding; go_function: magnesium ion binding; go_function: calcium-transporting ATPase activity; go_function: hydrolase activity; acting on acid anhydrides; catalyzing transmembrane movement of substances; go_process: metabolism; go_process: cation transport; go_process: proton transport; go_process: calcium ion transport; go_process: regulation of striated muscle contraction', 'Sarcolemmal-associated protein; go_component: integral to plasma membrane; go_component: smooth endoplasmic reticulum; go_function: unfolded protein binding; go_process: protein folding; go_process: muscle contraction', 'growth arrest-specific homeo box; go_component: nucleus; go_function: transcription factor activity; go_process: circulation; go_process: development; go_process: regulation of transcription; DNA-dependent', 'homeobox protein Hox-D3; Hox-4.1; mouse; homolog of; homeo box D3; go_component: nucleus; go_function: transcription factor activity; go_process: morphogenesis; go_process: regulation of transcription; DNA-dependent', 'isoform b is encoded by transcript variant 2; zinc finger DNA binding protein ZER6; zinc finger-estrogen receptor interaction; clone 6; zinc finger DNA binding protein p52/p71; go_component: nucleus; go_function: DNA binding; go_function: zinc ion binding; go_function: metal ion binding; go_function: transcriptional activator activity; go_process: transcription; go_process: regulation of transcription; DNA-dependent'], 'Product': ['ATPase; Ca++ transporting; fast twitch 1 isoform a', 'sarcolemma associated protein', 'mesenchyme homeo box 2', 'homeobox D3', 'zinc finger 398 isoform b'], 'Distance_to_TSS': [291.0, 369.0, 432.0, 268.0, 671.0], 'CPG_ISLAND': [True, True, True, False, True], 'CPG_ISLAND_LOCATIONS': ['16:28797486-28797825', '3:57716811-57718675', '7:15691512-15693551', nan, '7:148453584-148455804'], 'MIR_CPG_ISLAND': [nan, nan, nan, nan, nan], 'RANGE_GB': ['NC_000016.8', 'NC_000003.10', 'NC_000007.12', nan, 'NC_000007.12'], 'RANGE_START': [28797486.0, 57716811.0, 15691512.0, nan, 148453584.0], 'RANGE_END': [28797825.0, 57718675.0, 15693551.0, nan, 148455804.0], 'RANGE_STRAND': ['+', '+', '-', nan, '+'], 'GB_ACC': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'ORF': [487.0, 7871.0, 4223.0, 3232.0, 57541.0]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>IlmnStrand</th>\n",
       "      <th>AddressA_ID</th>\n",
       "      <th>AlleleA_ProbeSeq</th>\n",
       "      <th>AddressB_ID</th>\n",
       "      <th>AlleleB_ProbeSeq</th>\n",
       "      <th>GenomeBuild</th>\n",
       "      <th>Chr</th>\n",
       "      <th>MapInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>Distance_to_TSS</th>\n",
       "      <th>CPG_ISLAND</th>\n",
       "      <th>CPG_ISLAND_LOCATIONS</th>\n",
       "      <th>MIR_CPG_ISLAND</th>\n",
       "      <th>RANGE_GB</th>\n",
       "      <th>RANGE_START</th>\n",
       "      <th>RANGE_END</th>\n",
       "      <th>RANGE_STRAND</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>ORF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cg00000292</td>\n",
       "      <td>cg00000292</td>\n",
       "      <td>TOP</td>\n",
       "      <td>990370.0</td>\n",
       "      <td>AAACATTAATTACCAACCACTCTTCCAAAAAACACTTACCATTAAA...</td>\n",
       "      <td>6660678.0</td>\n",
       "      <td>AAACATTAATTACCAACCGCTCTTCCAAAAAACACTTACCATTAAA...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>16</td>\n",
       "      <td>28797601.0</td>\n",
       "      <td>...</td>\n",
       "      <td>291.0</td>\n",
       "      <td>True</td>\n",
       "      <td>16:28797486-28797825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000016.8</td>\n",
       "      <td>28797486.0</td>\n",
       "      <td>28797825.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_173201.2</td>\n",
       "      <td>487.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cg00002426</td>\n",
       "      <td>cg00002426</td>\n",
       "      <td>TOP</td>\n",
       "      <td>6580397.0</td>\n",
       "      <td>AATATAATAACATTACCTTACCCATCTTATAATCAAACCAAACAAA...</td>\n",
       "      <td>6100343.0</td>\n",
       "      <td>AATATAATAACATTACCTTACCCGTCTTATAATCAAACCAAACGAA...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>3</td>\n",
       "      <td>57718583.0</td>\n",
       "      <td>...</td>\n",
       "      <td>369.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3:57716811-57718675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000003.10</td>\n",
       "      <td>57716811.0</td>\n",
       "      <td>57718675.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_007159.2</td>\n",
       "      <td>7871.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cg00003994</td>\n",
       "      <td>cg00003994</td>\n",
       "      <td>TOP</td>\n",
       "      <td>7150184.0</td>\n",
       "      <td>AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTC...</td>\n",
       "      <td>7150392.0</td>\n",
       "      <td>AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTC...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>7</td>\n",
       "      <td>15692387.0</td>\n",
       "      <td>...</td>\n",
       "      <td>432.0</td>\n",
       "      <td>True</td>\n",
       "      <td>7:15691512-15693551</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000007.12</td>\n",
       "      <td>15691512.0</td>\n",
       "      <td>15693551.0</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_005924.3</td>\n",
       "      <td>4223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cg00005847</td>\n",
       "      <td>cg00005847</td>\n",
       "      <td>BOT</td>\n",
       "      <td>4850717.0</td>\n",
       "      <td>TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAA...</td>\n",
       "      <td>1260113.0</td>\n",
       "      <td>TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAA...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2</td>\n",
       "      <td>176737319.0</td>\n",
       "      <td>...</td>\n",
       "      <td>268.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_006898.4</td>\n",
       "      <td>3232.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cg00006414</td>\n",
       "      <td>cg00006414</td>\n",
       "      <td>BOT</td>\n",
       "      <td>6980731.0</td>\n",
       "      <td>CTCAAAAACCAAACAAAACAAAACCCCAATACTAATCATTAATAAA...</td>\n",
       "      <td>4280093.0</td>\n",
       "      <td>CTCGAAAACCGAACAAAACAAAACCCCAATACTAATCGTTAATAAA...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>7</td>\n",
       "      <td>148453770.0</td>\n",
       "      <td>...</td>\n",
       "      <td>671.0</td>\n",
       "      <td>True</td>\n",
       "      <td>7:148453584-148455804</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000007.12</td>\n",
       "      <td>148453584.0</td>\n",
       "      <td>148455804.0</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_020781.2</td>\n",
       "      <td>57541.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351365</th>\n",
       "      <td>cg27657283</td>\n",
       "      <td>0.1168531</td>\n",
       "      <td>3.68E-38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351366</th>\n",
       "      <td>cg27661264</td>\n",
       "      <td>0.3626379</td>\n",
       "      <td>3.68E-38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351367</th>\n",
       "      <td>cg27662379</td>\n",
       "      <td>0.03049343</td>\n",
       "      <td>3.68E-38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351368</th>\n",
       "      <td>cg27662877</td>\n",
       "      <td>0.05681526</td>\n",
       "      <td>3.68E-38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351369</th>\n",
       "      <td>cg27665659</td>\n",
       "      <td>0.04658268</td>\n",
       "      <td>3.68E-38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1351370 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID        Name IlmnStrand  AddressA_ID  \\\n",
       "0        cg00000292  cg00000292        TOP     990370.0   \n",
       "1        cg00002426  cg00002426        TOP    6580397.0   \n",
       "2        cg00003994  cg00003994        TOP    7150184.0   \n",
       "3        cg00005847  cg00005847        BOT    4850717.0   \n",
       "4        cg00006414  cg00006414        BOT    6980731.0   \n",
       "...             ...         ...        ...          ...   \n",
       "1351365  cg27657283   0.1168531   3.68E-38          NaN   \n",
       "1351366  cg27661264   0.3626379   3.68E-38          NaN   \n",
       "1351367  cg27662379  0.03049343   3.68E-38          NaN   \n",
       "1351368  cg27662877  0.05681526   3.68E-38          NaN   \n",
       "1351369  cg27665659  0.04658268   3.68E-38          NaN   \n",
       "\n",
       "                                          AlleleA_ProbeSeq  AddressB_ID  \\\n",
       "0        AAACATTAATTACCAACCACTCTTCCAAAAAACACTTACCATTAAA...    6660678.0   \n",
       "1        AATATAATAACATTACCTTACCCATCTTATAATCAAACCAAACAAA...    6100343.0   \n",
       "2        AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTC...    7150392.0   \n",
       "3        TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAA...    1260113.0   \n",
       "4        CTCAAAAACCAAACAAAACAAAACCCCAATACTAATCATTAATAAA...    4280093.0   \n",
       "...                                                    ...          ...   \n",
       "1351365                                                NaN          NaN   \n",
       "1351366                                                NaN          NaN   \n",
       "1351367                                                NaN          NaN   \n",
       "1351368                                                NaN          NaN   \n",
       "1351369                                                NaN          NaN   \n",
       "\n",
       "                                          AlleleB_ProbeSeq  GenomeBuild  Chr  \\\n",
       "0        AAACATTAATTACCAACCGCTCTTCCAAAAAACACTTACCATTAAA...         36.0   16   \n",
       "1        AATATAATAACATTACCTTACCCGTCTTATAATCAAACCAAACGAA...         36.0    3   \n",
       "2        AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTC...         36.0    7   \n",
       "3        TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAA...         36.0    2   \n",
       "4        CTCGAAAACCGAACAAAACAAAACCCCAATACTAATCGTTAATAAA...         36.0    7   \n",
       "...                                                    ...          ...  ...   \n",
       "1351365                                                NaN          NaN  NaN   \n",
       "1351366                                                NaN          NaN  NaN   \n",
       "1351367                                                NaN          NaN  NaN   \n",
       "1351368                                                NaN          NaN  NaN   \n",
       "1351369                                                NaN          NaN  NaN   \n",
       "\n",
       "             MapInfo  ... Distance_to_TSS CPG_ISLAND   CPG_ISLAND_LOCATIONS  \\\n",
       "0         28797601.0  ...           291.0       True   16:28797486-28797825   \n",
       "1         57718583.0  ...           369.0       True    3:57716811-57718675   \n",
       "2         15692387.0  ...           432.0       True    7:15691512-15693551   \n",
       "3        176737319.0  ...           268.0      False                    NaN   \n",
       "4        148453770.0  ...           671.0       True  7:148453584-148455804   \n",
       "...              ...  ...             ...        ...                    ...   \n",
       "1351365          NaN  ...             NaN        NaN                    NaN   \n",
       "1351366          NaN  ...             NaN        NaN                    NaN   \n",
       "1351367          NaN  ...             NaN        NaN                    NaN   \n",
       "1351368          NaN  ...             NaN        NaN                    NaN   \n",
       "1351369          NaN  ...             NaN        NaN                    NaN   \n",
       "\n",
       "         MIR_CPG_ISLAND      RANGE_GB  RANGE_START    RANGE_END RANGE_STRAND  \\\n",
       "0                   NaN   NC_000016.8   28797486.0   28797825.0            +   \n",
       "1                   NaN  NC_000003.10   57716811.0   57718675.0            +   \n",
       "2                   NaN  NC_000007.12   15691512.0   15693551.0            -   \n",
       "3                   NaN           NaN          NaN          NaN          NaN   \n",
       "4                   NaN  NC_000007.12  148453584.0  148455804.0            +   \n",
       "...                 ...           ...          ...          ...          ...   \n",
       "1351365             NaN           NaN          NaN          NaN          NaN   \n",
       "1351366             NaN           NaN          NaN          NaN          NaN   \n",
       "1351367             NaN           NaN          NaN          NaN          NaN   \n",
       "1351368             NaN           NaN          NaN          NaN          NaN   \n",
       "1351369             NaN           NaN          NaN          NaN          NaN   \n",
       "\n",
       "              GB_ACC      ORF  \n",
       "0        NM_173201.2    487.0  \n",
       "1        NM_007159.2   7871.0  \n",
       "2        NM_005924.3   4223.0  \n",
       "3        NM_006898.4   3232.0  \n",
       "4        NM_020781.2  57541.0  \n",
       "...              ...      ...  \n",
       "1351365          NaN      NaN  \n",
       "1351366          NaN      NaN  \n",
       "1351367          NaN      NaN  \n",
       "1351368          NaN      NaN  \n",
       "1351369          NaN      NaN  \n",
       "\n",
       "[1351370 rows x 38 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Name', 'IlmnStrand', 'AddressA_ID', 'AlleleA_ProbeSeq',\n",
       "       'AddressB_ID', 'AlleleB_ProbeSeq', 'GenomeBuild', 'Chr', 'MapInfo',\n",
       "       'Ploidy', 'Species', 'Source', 'SourceVersion', 'SourceStrand',\n",
       "       'SourceSeq', 'TopGenomicSeq', 'Next_Base', 'Color_Channel',\n",
       "       'TSS_Coordinate', 'Gene_Strand', 'Gene_ID', 'Symbol', 'Synonym',\n",
       "       'Accession', 'GID', 'Annotation', 'Product', 'Distance_to_TSS',\n",
       "       'CPG_ISLAND', 'CPG_ISLAND_LOCATIONS', 'MIR_CPG_ISLAND', 'RANGE_GB',\n",
       "       'RANGE_START', 'RANGE_END', 'RANGE_STRAND', 'GB_ACC', 'ORF'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          ATP2A1\n",
       "1           SLMAP\n",
       "2           MEOX2\n",
       "3           HOXD3\n",
       "4          ZNF398\n",
       "            ...  \n",
       "1351365       NaN\n",
       "1351366       NaN\n",
       "1351367       NaN\n",
       "1351368       NaN\n",
       "1351369       NaN\n",
       "Name: Symbol, Length: 1351370, dtype: object"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation['Symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM889444</th>\n",
       "      <th>GSM889445</th>\n",
       "      <th>GSM889446</th>\n",
       "      <th>GSM889447</th>\n",
       "      <th>GSM889448</th>\n",
       "      <th>GSM889449</th>\n",
       "      <th>GSM889450</th>\n",
       "      <th>GSM889451</th>\n",
       "      <th>GSM889452</th>\n",
       "      <th>GSM889453</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM889482</th>\n",
       "      <th>GSM889483</th>\n",
       "      <th>GSM889484</th>\n",
       "      <th>GSM889485</th>\n",
       "      <th>GSM889486</th>\n",
       "      <th>GSM889487</th>\n",
       "      <th>GSM889488</th>\n",
       "      <th>GSM889489</th>\n",
       "      <th>GSM889490</th>\n",
       "      <th>GSM889491</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39873</th>\n",
       "      <td>0.689279</td>\n",
       "      <td>0.426369</td>\n",
       "      <td>0.711409</td>\n",
       "      <td>0.107636</td>\n",
       "      <td>0.801374</td>\n",
       "      <td>0.333542</td>\n",
       "      <td>0.485201</td>\n",
       "      <td>0.689080</td>\n",
       "      <td>0.308070</td>\n",
       "      <td>0.429268</td>\n",
       "      <td>...</td>\n",
       "      <td>0.551247</td>\n",
       "      <td>0.516344</td>\n",
       "      <td>0.667079</td>\n",
       "      <td>0.695478</td>\n",
       "      <td>0.610036</td>\n",
       "      <td>0.669368</td>\n",
       "      <td>0.729632</td>\n",
       "      <td>0.660225</td>\n",
       "      <td>0.679691</td>\n",
       "      <td>0.660214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39874</th>\n",
       "      <td>0.064024</td>\n",
       "      <td>0.052674</td>\n",
       "      <td>0.061257</td>\n",
       "      <td>0.056175</td>\n",
       "      <td>0.056825</td>\n",
       "      <td>0.041682</td>\n",
       "      <td>0.082841</td>\n",
       "      <td>0.067413</td>\n",
       "      <td>0.070280</td>\n",
       "      <td>0.070044</td>\n",
       "      <td>...</td>\n",
       "      <td>0.114598</td>\n",
       "      <td>0.101350</td>\n",
       "      <td>0.094201</td>\n",
       "      <td>0.076669</td>\n",
       "      <td>0.067392</td>\n",
       "      <td>0.081730</td>\n",
       "      <td>0.053490</td>\n",
       "      <td>0.067670</td>\n",
       "      <td>0.055647</td>\n",
       "      <td>0.062393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39875</th>\n",
       "      <td>0.249730</td>\n",
       "      <td>0.438115</td>\n",
       "      <td>0.440215</td>\n",
       "      <td>0.441576</td>\n",
       "      <td>0.441734</td>\n",
       "      <td>0.247551</td>\n",
       "      <td>0.368339</td>\n",
       "      <td>0.409045</td>\n",
       "      <td>0.252633</td>\n",
       "      <td>0.203710</td>\n",
       "      <td>...</td>\n",
       "      <td>0.388586</td>\n",
       "      <td>0.430048</td>\n",
       "      <td>0.276230</td>\n",
       "      <td>0.268600</td>\n",
       "      <td>0.387995</td>\n",
       "      <td>0.307379</td>\n",
       "      <td>0.350970</td>\n",
       "      <td>0.283600</td>\n",
       "      <td>0.330482</td>\n",
       "      <td>0.310958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39877</th>\n",
       "      <td>0.057814</td>\n",
       "      <td>0.059389</td>\n",
       "      <td>0.070888</td>\n",
       "      <td>0.052026</td>\n",
       "      <td>0.048152</td>\n",
       "      <td>0.061054</td>\n",
       "      <td>0.091866</td>\n",
       "      <td>0.059028</td>\n",
       "      <td>0.071877</td>\n",
       "      <td>0.055394</td>\n",
       "      <td>...</td>\n",
       "      <td>0.092821</td>\n",
       "      <td>0.075314</td>\n",
       "      <td>0.072169</td>\n",
       "      <td>0.057845</td>\n",
       "      <td>0.062215</td>\n",
       "      <td>0.070240</td>\n",
       "      <td>0.045412</td>\n",
       "      <td>0.056299</td>\n",
       "      <td>0.053210</td>\n",
       "      <td>0.069765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39878</th>\n",
       "      <td>0.049203</td>\n",
       "      <td>0.059566</td>\n",
       "      <td>0.053505</td>\n",
       "      <td>0.048289</td>\n",
       "      <td>0.046017</td>\n",
       "      <td>0.044724</td>\n",
       "      <td>0.060468</td>\n",
       "      <td>0.046377</td>\n",
       "      <td>0.058192</td>\n",
       "      <td>0.047501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077212</td>\n",
       "      <td>0.072336</td>\n",
       "      <td>0.057878</td>\n",
       "      <td>0.049363</td>\n",
       "      <td>0.055631</td>\n",
       "      <td>0.059790</td>\n",
       "      <td>0.053111</td>\n",
       "      <td>0.054395</td>\n",
       "      <td>0.054666</td>\n",
       "      <td>0.049155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-D3</th>\n",
       "      <td>0.135123</td>\n",
       "      <td>0.167544</td>\n",
       "      <td>0.120367</td>\n",
       "      <td>0.074781</td>\n",
       "      <td>0.092117</td>\n",
       "      <td>0.226392</td>\n",
       "      <td>0.219823</td>\n",
       "      <td>0.273371</td>\n",
       "      <td>0.114962</td>\n",
       "      <td>0.117908</td>\n",
       "      <td>...</td>\n",
       "      <td>0.220372</td>\n",
       "      <td>0.249601</td>\n",
       "      <td>0.157433</td>\n",
       "      <td>0.137902</td>\n",
       "      <td>0.187125</td>\n",
       "      <td>0.139870</td>\n",
       "      <td>0.175214</td>\n",
       "      <td>0.196041</td>\n",
       "      <td>0.176399</td>\n",
       "      <td>0.191310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-H2</th>\n",
       "      <td>0.031693</td>\n",
       "      <td>0.037456</td>\n",
       "      <td>0.034644</td>\n",
       "      <td>0.031599</td>\n",
       "      <td>0.040484</td>\n",
       "      <td>0.028122</td>\n",
       "      <td>0.030194</td>\n",
       "      <td>0.026827</td>\n",
       "      <td>0.049008</td>\n",
       "      <td>0.043003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035374</td>\n",
       "      <td>0.036672</td>\n",
       "      <td>0.036632</td>\n",
       "      <td>0.036060</td>\n",
       "      <td>0.035811</td>\n",
       "      <td>0.034073</td>\n",
       "      <td>0.027685</td>\n",
       "      <td>0.033319</td>\n",
       "      <td>0.031560</td>\n",
       "      <td>0.029071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hfl-B5</th>\n",
       "      <td>0.031899</td>\n",
       "      <td>0.034486</td>\n",
       "      <td>0.035470</td>\n",
       "      <td>0.036843</td>\n",
       "      <td>0.034103</td>\n",
       "      <td>0.027757</td>\n",
       "      <td>0.037554</td>\n",
       "      <td>0.037559</td>\n",
       "      <td>0.037354</td>\n",
       "      <td>0.038608</td>\n",
       "      <td>...</td>\n",
       "      <td>0.036248</td>\n",
       "      <td>0.041694</td>\n",
       "      <td>0.043964</td>\n",
       "      <td>0.044827</td>\n",
       "      <td>0.039061</td>\n",
       "      <td>0.038787</td>\n",
       "      <td>0.036774</td>\n",
       "      <td>0.039394</td>\n",
       "      <td>0.043053</td>\n",
       "      <td>0.032806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mimitin</th>\n",
       "      <td>0.660241</td>\n",
       "      <td>0.706996</td>\n",
       "      <td>0.743767</td>\n",
       "      <td>0.604390</td>\n",
       "      <td>0.677875</td>\n",
       "      <td>0.775538</td>\n",
       "      <td>0.397100</td>\n",
       "      <td>0.621684</td>\n",
       "      <td>0.769524</td>\n",
       "      <td>0.732552</td>\n",
       "      <td>...</td>\n",
       "      <td>0.562751</td>\n",
       "      <td>0.561515</td>\n",
       "      <td>0.598229</td>\n",
       "      <td>0.598634</td>\n",
       "      <td>0.543056</td>\n",
       "      <td>0.612676</td>\n",
       "      <td>0.617838</td>\n",
       "      <td>0.579518</td>\n",
       "      <td>0.578914</td>\n",
       "      <td>0.560829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>raptor</th>\n",
       "      <td>0.073761</td>\n",
       "      <td>0.061756</td>\n",
       "      <td>0.091470</td>\n",
       "      <td>0.053681</td>\n",
       "      <td>0.097462</td>\n",
       "      <td>0.090294</td>\n",
       "      <td>0.144359</td>\n",
       "      <td>0.109272</td>\n",
       "      <td>0.126251</td>\n",
       "      <td>0.095938</td>\n",
       "      <td>...</td>\n",
       "      <td>0.120534</td>\n",
       "      <td>0.101582</td>\n",
       "      <td>0.139143</td>\n",
       "      <td>0.085799</td>\n",
       "      <td>0.087462</td>\n",
       "      <td>0.092672</td>\n",
       "      <td>0.062942</td>\n",
       "      <td>0.108999</td>\n",
       "      <td>0.079347</td>\n",
       "      <td>0.086888</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14436 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM889444  GSM889445  GSM889446  GSM889447  GSM889448  GSM889449  \\\n",
       "Gene                                                                        \n",
       "39873     0.689279   0.426369   0.711409   0.107636   0.801374   0.333542   \n",
       "39874     0.064024   0.052674   0.061257   0.056175   0.056825   0.041682   \n",
       "39875     0.249730   0.438115   0.440215   0.441576   0.441734   0.247551   \n",
       "39877     0.057814   0.059389   0.070888   0.052026   0.048152   0.061054   \n",
       "39878     0.049203   0.059566   0.053505   0.048289   0.046017   0.044724   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "hCAP-D3   0.135123   0.167544   0.120367   0.074781   0.092117   0.226392   \n",
       "hCAP-H2   0.031693   0.037456   0.034644   0.031599   0.040484   0.028122   \n",
       "hfl-B5    0.031899   0.034486   0.035470   0.036843   0.034103   0.027757   \n",
       "mimitin   0.660241   0.706996   0.743767   0.604390   0.677875   0.775538   \n",
       "raptor    0.073761   0.061756   0.091470   0.053681   0.097462   0.090294   \n",
       "\n",
       "         GSM889450  GSM889451  GSM889452  GSM889453  ...  GSM889482  \\\n",
       "Gene                                                 ...              \n",
       "39873     0.485201   0.689080   0.308070   0.429268  ...   0.551247   \n",
       "39874     0.082841   0.067413   0.070280   0.070044  ...   0.114598   \n",
       "39875     0.368339   0.409045   0.252633   0.203710  ...   0.388586   \n",
       "39877     0.091866   0.059028   0.071877   0.055394  ...   0.092821   \n",
       "39878     0.060468   0.046377   0.058192   0.047501  ...   0.077212   \n",
       "...            ...        ...        ...        ...  ...        ...   \n",
       "hCAP-D3   0.219823   0.273371   0.114962   0.117908  ...   0.220372   \n",
       "hCAP-H2   0.030194   0.026827   0.049008   0.043003  ...   0.035374   \n",
       "hfl-B5    0.037554   0.037559   0.037354   0.038608  ...   0.036248   \n",
       "mimitin   0.397100   0.621684   0.769524   0.732552  ...   0.562751   \n",
       "raptor    0.144359   0.109272   0.126251   0.095938  ...   0.120534   \n",
       "\n",
       "         GSM889483  GSM889484  GSM889485  GSM889486  GSM889487  GSM889488  \\\n",
       "Gene                                                                        \n",
       "39873     0.516344   0.667079   0.695478   0.610036   0.669368   0.729632   \n",
       "39874     0.101350   0.094201   0.076669   0.067392   0.081730   0.053490   \n",
       "39875     0.430048   0.276230   0.268600   0.387995   0.307379   0.350970   \n",
       "39877     0.075314   0.072169   0.057845   0.062215   0.070240   0.045412   \n",
       "39878     0.072336   0.057878   0.049363   0.055631   0.059790   0.053111   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "hCAP-D3   0.249601   0.157433   0.137902   0.187125   0.139870   0.175214   \n",
       "hCAP-H2   0.036672   0.036632   0.036060   0.035811   0.034073   0.027685   \n",
       "hfl-B5    0.041694   0.043964   0.044827   0.039061   0.038787   0.036774   \n",
       "mimitin   0.561515   0.598229   0.598634   0.543056   0.612676   0.617838   \n",
       "raptor    0.101582   0.139143   0.085799   0.087462   0.092672   0.062942   \n",
       "\n",
       "         GSM889489  GSM889490  GSM889491  \n",
       "Gene                                      \n",
       "39873     0.660225   0.679691   0.660214  \n",
       "39874     0.067670   0.055647   0.062393  \n",
       "39875     0.283600   0.330482   0.310958  \n",
       "39877     0.056299   0.053210   0.069765  \n",
       "39878     0.054395   0.054666   0.049155  \n",
       "...            ...        ...        ...  \n",
       "hCAP-D3   0.196041   0.176399   0.191310  \n",
       "hCAP-H2   0.033319   0.031560   0.029071  \n",
       "hfl-B5    0.039394   0.043053   0.032806  \n",
       "mimitin   0.579518   0.578914   0.560829  \n",
       "raptor    0.108999   0.079347   0.086888  \n",
       "\n",
       "[14436 rows x 48 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM889444</th>\n",
       "      <th>GSM889445</th>\n",
       "      <th>GSM889446</th>\n",
       "      <th>GSM889447</th>\n",
       "      <th>GSM889448</th>\n",
       "      <th>GSM889449</th>\n",
       "      <th>GSM889450</th>\n",
       "      <th>GSM889451</th>\n",
       "      <th>GSM889452</th>\n",
       "      <th>GSM889453</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM889482</th>\n",
       "      <th>GSM889483</th>\n",
       "      <th>GSM889484</th>\n",
       "      <th>GSM889485</th>\n",
       "      <th>GSM889486</th>\n",
       "      <th>GSM889487</th>\n",
       "      <th>GSM889488</th>\n",
       "      <th>GSM889489</th>\n",
       "      <th>GSM889490</th>\n",
       "      <th>GSM889491</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.877513</td>\n",
       "      <td>0.790216</td>\n",
       "      <td>0.905854</td>\n",
       "      <td>0.926305</td>\n",
       "      <td>0.813567</td>\n",
       "      <td>0.935604</td>\n",
       "      <td>0.876431</td>\n",
       "      <td>0.900982</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.916555</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890737</td>\n",
       "      <td>0.775418</td>\n",
       "      <td>0.609464</td>\n",
       "      <td>0.909835</td>\n",
       "      <td>0.908542</td>\n",
       "      <td>0.770064</td>\n",
       "      <td>0.901580</td>\n",
       "      <td>0.836649</td>\n",
       "      <td>0.881275</td>\n",
       "      <td>0.840217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.488576</td>\n",
       "      <td>0.616612</td>\n",
       "      <td>0.415186</td>\n",
       "      <td>0.450625</td>\n",
       "      <td>0.524828</td>\n",
       "      <td>0.255921</td>\n",
       "      <td>0.446866</td>\n",
       "      <td>0.419583</td>\n",
       "      <td>0.255085</td>\n",
       "      <td>0.402749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.427695</td>\n",
       "      <td>0.473264</td>\n",
       "      <td>0.414352</td>\n",
       "      <td>0.301115</td>\n",
       "      <td>0.429467</td>\n",
       "      <td>0.366102</td>\n",
       "      <td>0.558921</td>\n",
       "      <td>0.368966</td>\n",
       "      <td>0.398305</td>\n",
       "      <td>0.449060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.770778</td>\n",
       "      <td>0.822486</td>\n",
       "      <td>0.758853</td>\n",
       "      <td>0.844311</td>\n",
       "      <td>0.796473</td>\n",
       "      <td>0.622104</td>\n",
       "      <td>0.740103</td>\n",
       "      <td>0.874109</td>\n",
       "      <td>0.719518</td>\n",
       "      <td>0.821388</td>\n",
       "      <td>...</td>\n",
       "      <td>0.547426</td>\n",
       "      <td>0.732379</td>\n",
       "      <td>0.728085</td>\n",
       "      <td>0.755285</td>\n",
       "      <td>0.723343</td>\n",
       "      <td>0.746495</td>\n",
       "      <td>0.798775</td>\n",
       "      <td>0.800428</td>\n",
       "      <td>0.767000</td>\n",
       "      <td>0.759809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.337080</td>\n",
       "      <td>0.623585</td>\n",
       "      <td>0.400251</td>\n",
       "      <td>0.422990</td>\n",
       "      <td>0.342717</td>\n",
       "      <td>0.503112</td>\n",
       "      <td>0.257814</td>\n",
       "      <td>0.359270</td>\n",
       "      <td>0.445166</td>\n",
       "      <td>0.171016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.306420</td>\n",
       "      <td>0.382156</td>\n",
       "      <td>0.371309</td>\n",
       "      <td>0.335879</td>\n",
       "      <td>0.314983</td>\n",
       "      <td>0.302081</td>\n",
       "      <td>0.317394</td>\n",
       "      <td>0.357834</td>\n",
       "      <td>0.334070</td>\n",
       "      <td>0.320748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>0.731916</td>\n",
       "      <td>0.760295</td>\n",
       "      <td>0.797876</td>\n",
       "      <td>0.610334</td>\n",
       "      <td>0.804778</td>\n",
       "      <td>0.720754</td>\n",
       "      <td>0.722959</td>\n",
       "      <td>0.862955</td>\n",
       "      <td>0.590652</td>\n",
       "      <td>0.819226</td>\n",
       "      <td>...</td>\n",
       "      <td>0.646292</td>\n",
       "      <td>0.768062</td>\n",
       "      <td>0.742481</td>\n",
       "      <td>0.785946</td>\n",
       "      <td>0.809538</td>\n",
       "      <td>0.741358</td>\n",
       "      <td>0.797128</td>\n",
       "      <td>0.759670</td>\n",
       "      <td>0.774546</td>\n",
       "      <td>0.740727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWINT</th>\n",
       "      <td>0.055689</td>\n",
       "      <td>0.054493</td>\n",
       "      <td>0.046531</td>\n",
       "      <td>0.050910</td>\n",
       "      <td>0.045079</td>\n",
       "      <td>0.039365</td>\n",
       "      <td>0.099023</td>\n",
       "      <td>0.051996</td>\n",
       "      <td>0.063345</td>\n",
       "      <td>0.051017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058510</td>\n",
       "      <td>0.056668</td>\n",
       "      <td>0.048632</td>\n",
       "      <td>0.050794</td>\n",
       "      <td>0.049920</td>\n",
       "      <td>0.060809</td>\n",
       "      <td>0.050605</td>\n",
       "      <td>0.049655</td>\n",
       "      <td>0.052283</td>\n",
       "      <td>0.048284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDA</th>\n",
       "      <td>0.471868</td>\n",
       "      <td>0.394791</td>\n",
       "      <td>0.417707</td>\n",
       "      <td>0.432613</td>\n",
       "      <td>0.440079</td>\n",
       "      <td>0.430349</td>\n",
       "      <td>0.407978</td>\n",
       "      <td>0.394574</td>\n",
       "      <td>0.462678</td>\n",
       "      <td>0.421530</td>\n",
       "      <td>...</td>\n",
       "      <td>0.346426</td>\n",
       "      <td>0.405079</td>\n",
       "      <td>0.447421</td>\n",
       "      <td>0.421397</td>\n",
       "      <td>0.404630</td>\n",
       "      <td>0.384359</td>\n",
       "      <td>0.463203</td>\n",
       "      <td>0.400643</td>\n",
       "      <td>0.387745</td>\n",
       "      <td>0.397563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.028720</td>\n",
       "      <td>0.037781</td>\n",
       "      <td>0.039477</td>\n",
       "      <td>0.034675</td>\n",
       "      <td>0.032816</td>\n",
       "      <td>0.038400</td>\n",
       "      <td>0.133074</td>\n",
       "      <td>0.034445</td>\n",
       "      <td>0.042211</td>\n",
       "      <td>0.035093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.061896</td>\n",
       "      <td>0.050344</td>\n",
       "      <td>0.037829</td>\n",
       "      <td>0.034190</td>\n",
       "      <td>0.034174</td>\n",
       "      <td>0.049541</td>\n",
       "      <td>0.040220</td>\n",
       "      <td>0.030688</td>\n",
       "      <td>0.031048</td>\n",
       "      <td>0.036458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.080601</td>\n",
       "      <td>0.086067</td>\n",
       "      <td>0.051746</td>\n",
       "      <td>0.062786</td>\n",
       "      <td>0.406175</td>\n",
       "      <td>0.085561</td>\n",
       "      <td>0.072221</td>\n",
       "      <td>0.066332</td>\n",
       "      <td>0.064198</td>\n",
       "      <td>0.065377</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100295</td>\n",
       "      <td>0.118933</td>\n",
       "      <td>0.086164</td>\n",
       "      <td>0.100369</td>\n",
       "      <td>0.125893</td>\n",
       "      <td>0.076795</td>\n",
       "      <td>0.083921</td>\n",
       "      <td>0.100283</td>\n",
       "      <td>0.091715</td>\n",
       "      <td>0.083698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.050376</td>\n",
       "      <td>0.049650</td>\n",
       "      <td>0.061866</td>\n",
       "      <td>0.061720</td>\n",
       "      <td>0.057205</td>\n",
       "      <td>0.049556</td>\n",
       "      <td>0.104789</td>\n",
       "      <td>0.054134</td>\n",
       "      <td>0.061472</td>\n",
       "      <td>0.068653</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084454</td>\n",
       "      <td>0.069341</td>\n",
       "      <td>0.069776</td>\n",
       "      <td>0.055138</td>\n",
       "      <td>0.054799</td>\n",
       "      <td>0.066625</td>\n",
       "      <td>0.049213</td>\n",
       "      <td>0.054959</td>\n",
       "      <td>0.047587</td>\n",
       "      <td>0.055029</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11144 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM889444  GSM889445  GSM889446  GSM889447  GSM889448  GSM889449  \\\n",
       "A1BG     0.877513   0.790216   0.905854   0.926305   0.813567   0.935604   \n",
       "A2M      0.488576   0.616612   0.415186   0.450625   0.524828   0.255921   \n",
       "A2ML1    0.770778   0.822486   0.758853   0.844311   0.796473   0.622104   \n",
       "A4GALT   0.337080   0.623585   0.400251   0.422990   0.342717   0.503112   \n",
       "A4GNT    0.731916   0.760295   0.797876   0.610334   0.804778   0.720754   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZWINT    0.055689   0.054493   0.046531   0.050910   0.045079   0.039365   \n",
       "ZXDA     0.471868   0.394791   0.417707   0.432613   0.440079   0.430349   \n",
       "ZYX      0.028720   0.037781   0.039477   0.034675   0.032816   0.038400   \n",
       "ZZEF1    0.080601   0.086067   0.051746   0.062786   0.406175   0.085561   \n",
       "ZZZ3     0.050376   0.049650   0.061866   0.061720   0.057205   0.049556   \n",
       "\n",
       "        GSM889450  GSM889451  GSM889452  GSM889453  ...  GSM889482  GSM889483  \\\n",
       "A1BG     0.876431   0.900982   0.923077   0.916555  ...   0.890737   0.775418   \n",
       "A2M      0.446866   0.419583   0.255085   0.402749  ...   0.427695   0.473264   \n",
       "A2ML1    0.740103   0.874109   0.719518   0.821388  ...   0.547426   0.732379   \n",
       "A4GALT   0.257814   0.359270   0.445166   0.171016  ...   0.306420   0.382156   \n",
       "A4GNT    0.722959   0.862955   0.590652   0.819226  ...   0.646292   0.768062   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "ZWINT    0.099023   0.051996   0.063345   0.051017  ...   0.058510   0.056668   \n",
       "ZXDA     0.407978   0.394574   0.462678   0.421530  ...   0.346426   0.405079   \n",
       "ZYX      0.133074   0.034445   0.042211   0.035093  ...   0.061896   0.050344   \n",
       "ZZEF1    0.072221   0.066332   0.064198   0.065377  ...   0.100295   0.118933   \n",
       "ZZZ3     0.104789   0.054134   0.061472   0.068653  ...   0.084454   0.069341   \n",
       "\n",
       "        GSM889484  GSM889485  GSM889486  GSM889487  GSM889488  GSM889489  \\\n",
       "A1BG     0.609464   0.909835   0.908542   0.770064   0.901580   0.836649   \n",
       "A2M      0.414352   0.301115   0.429467   0.366102   0.558921   0.368966   \n",
       "A2ML1    0.728085   0.755285   0.723343   0.746495   0.798775   0.800428   \n",
       "A4GALT   0.371309   0.335879   0.314983   0.302081   0.317394   0.357834   \n",
       "A4GNT    0.742481   0.785946   0.809538   0.741358   0.797128   0.759670   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZWINT    0.048632   0.050794   0.049920   0.060809   0.050605   0.049655   \n",
       "ZXDA     0.447421   0.421397   0.404630   0.384359   0.463203   0.400643   \n",
       "ZYX      0.037829   0.034190   0.034174   0.049541   0.040220   0.030688   \n",
       "ZZEF1    0.086164   0.100369   0.125893   0.076795   0.083921   0.100283   \n",
       "ZZZ3     0.069776   0.055138   0.054799   0.066625   0.049213   0.054959   \n",
       "\n",
       "        GSM889490  GSM889491  \n",
       "A1BG     0.881275   0.840217  \n",
       "A2M      0.398305   0.449060  \n",
       "A2ML1    0.767000   0.759809  \n",
       "A4GALT   0.334070   0.320748  \n",
       "A4GNT    0.774546   0.740727  \n",
       "...           ...        ...  \n",
       "ZWINT    0.052283   0.048284  \n",
       "ZXDA     0.387745   0.397563  \n",
       "ZYX      0.031048   0.036458  \n",
       "ZZEF1    0.091715   0.083698  \n",
       "ZZZ3     0.047587   0.055029  \n",
       "\n",
       "[11144 rows x 48 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease state: normal adrenal tissue':\n",
    "        return 0  # Epilepsy present\n",
    "    else:\n",
    "        return 1  # Epilepsy not present\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM889444</th>\n",
       "      <th>GSM889445</th>\n",
       "      <th>GSM889446</th>\n",
       "      <th>GSM889447</th>\n",
       "      <th>GSM889448</th>\n",
       "      <th>GSM889449</th>\n",
       "      <th>GSM889450</th>\n",
       "      <th>GSM889451</th>\n",
       "      <th>GSM889452</th>\n",
       "      <th>GSM889453</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM889482</th>\n",
       "      <th>GSM889483</th>\n",
       "      <th>GSM889484</th>\n",
       "      <th>GSM889485</th>\n",
       "      <th>GSM889486</th>\n",
       "      <th>GSM889487</th>\n",
       "      <th>GSM889488</th>\n",
       "      <th>GSM889489</th>\n",
       "      <th>GSM889490</th>\n",
       "      <th>GSM889491</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM889444  GSM889445  GSM889446  GSM889447  GSM889448  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "Gender                         1          1          1          1          1   \n",
       "\n",
       "                       GSM889449  GSM889450  GSM889451  GSM889452  GSM889453  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "Gender                         0          1          0          0          1   \n",
       "\n",
       "                       ...  GSM889482  GSM889483  GSM889484  GSM889485  \\\n",
       "Adrenocortical Cancer  ...          1          1          1          1   \n",
       "Gender                 ...          1          1          1          0   \n",
       "\n",
       "                       GSM889486  GSM889487  GSM889488  GSM889489  GSM889490  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "Gender                         0          1          0          1          0   \n",
       "\n",
       "                       GSM889491  \n",
       "Adrenocortical Cancer          0  \n",
       "Gender                         1  \n",
       "\n",
       "[2 rows x 48 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN2</th>\n",
       "      <th>ZSCAN4</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM889444</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.877513</td>\n",
       "      <td>0.488576</td>\n",
       "      <td>0.770778</td>\n",
       "      <td>0.337080</td>\n",
       "      <td>0.731916</td>\n",
       "      <td>0.152761</td>\n",
       "      <td>0.145945</td>\n",
       "      <td>0.883583</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060651</td>\n",
       "      <td>0.861556</td>\n",
       "      <td>0.113921</td>\n",
       "      <td>0.039714</td>\n",
       "      <td>0.068936</td>\n",
       "      <td>0.055689</td>\n",
       "      <td>0.471868</td>\n",
       "      <td>0.028720</td>\n",
       "      <td>0.080601</td>\n",
       "      <td>0.050376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889445</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.790216</td>\n",
       "      <td>0.616612</td>\n",
       "      <td>0.822486</td>\n",
       "      <td>0.623585</td>\n",
       "      <td>0.760295</td>\n",
       "      <td>0.155780</td>\n",
       "      <td>0.222403</td>\n",
       "      <td>0.916509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078385</td>\n",
       "      <td>0.597086</td>\n",
       "      <td>0.099773</td>\n",
       "      <td>0.044380</td>\n",
       "      <td>0.077357</td>\n",
       "      <td>0.054493</td>\n",
       "      <td>0.394791</td>\n",
       "      <td>0.037781</td>\n",
       "      <td>0.086067</td>\n",
       "      <td>0.049650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889446</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.905854</td>\n",
       "      <td>0.415186</td>\n",
       "      <td>0.758853</td>\n",
       "      <td>0.400251</td>\n",
       "      <td>0.797876</td>\n",
       "      <td>0.135304</td>\n",
       "      <td>0.105125</td>\n",
       "      <td>0.853768</td>\n",
       "      <td>...</td>\n",
       "      <td>0.090061</td>\n",
       "      <td>0.935782</td>\n",
       "      <td>0.100392</td>\n",
       "      <td>0.050342</td>\n",
       "      <td>0.099744</td>\n",
       "      <td>0.046531</td>\n",
       "      <td>0.417707</td>\n",
       "      <td>0.039477</td>\n",
       "      <td>0.051746</td>\n",
       "      <td>0.061866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889447</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.926305</td>\n",
       "      <td>0.450625</td>\n",
       "      <td>0.844311</td>\n",
       "      <td>0.422990</td>\n",
       "      <td>0.610334</td>\n",
       "      <td>0.123305</td>\n",
       "      <td>0.076724</td>\n",
       "      <td>0.693930</td>\n",
       "      <td>...</td>\n",
       "      <td>0.061925</td>\n",
       "      <td>0.113502</td>\n",
       "      <td>0.082969</td>\n",
       "      <td>0.042907</td>\n",
       "      <td>0.059283</td>\n",
       "      <td>0.050910</td>\n",
       "      <td>0.432613</td>\n",
       "      <td>0.034675</td>\n",
       "      <td>0.062786</td>\n",
       "      <td>0.061720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889448</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.813567</td>\n",
       "      <td>0.524828</td>\n",
       "      <td>0.796473</td>\n",
       "      <td>0.342717</td>\n",
       "      <td>0.804778</td>\n",
       "      <td>0.091906</td>\n",
       "      <td>0.106987</td>\n",
       "      <td>0.864323</td>\n",
       "      <td>...</td>\n",
       "      <td>0.072116</td>\n",
       "      <td>0.926427</td>\n",
       "      <td>0.096655</td>\n",
       "      <td>0.049146</td>\n",
       "      <td>0.077697</td>\n",
       "      <td>0.045079</td>\n",
       "      <td>0.440079</td>\n",
       "      <td>0.032816</td>\n",
       "      <td>0.406175</td>\n",
       "      <td>0.057205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889449</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.935604</td>\n",
       "      <td>0.255921</td>\n",
       "      <td>0.622104</td>\n",
       "      <td>0.503112</td>\n",
       "      <td>0.720754</td>\n",
       "      <td>0.104102</td>\n",
       "      <td>0.098242</td>\n",
       "      <td>0.771726</td>\n",
       "      <td>...</td>\n",
       "      <td>0.051829</td>\n",
       "      <td>0.922562</td>\n",
       "      <td>0.070781</td>\n",
       "      <td>0.041381</td>\n",
       "      <td>0.067504</td>\n",
       "      <td>0.039365</td>\n",
       "      <td>0.430349</td>\n",
       "      <td>0.038400</td>\n",
       "      <td>0.085561</td>\n",
       "      <td>0.049556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889450</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.876431</td>\n",
       "      <td>0.446866</td>\n",
       "      <td>0.740103</td>\n",
       "      <td>0.257814</td>\n",
       "      <td>0.722959</td>\n",
       "      <td>0.163676</td>\n",
       "      <td>0.144049</td>\n",
       "      <td>0.902074</td>\n",
       "      <td>...</td>\n",
       "      <td>0.101449</td>\n",
       "      <td>0.865154</td>\n",
       "      <td>0.165601</td>\n",
       "      <td>0.050697</td>\n",
       "      <td>0.126225</td>\n",
       "      <td>0.099023</td>\n",
       "      <td>0.407978</td>\n",
       "      <td>0.133074</td>\n",
       "      <td>0.072221</td>\n",
       "      <td>0.104789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889451</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.900982</td>\n",
       "      <td>0.419583</td>\n",
       "      <td>0.874109</td>\n",
       "      <td>0.359270</td>\n",
       "      <td>0.862955</td>\n",
       "      <td>0.117089</td>\n",
       "      <td>0.104022</td>\n",
       "      <td>0.932222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076846</td>\n",
       "      <td>0.939495</td>\n",
       "      <td>0.100747</td>\n",
       "      <td>0.043479</td>\n",
       "      <td>0.089807</td>\n",
       "      <td>0.051996</td>\n",
       "      <td>0.394574</td>\n",
       "      <td>0.034445</td>\n",
       "      <td>0.066332</td>\n",
       "      <td>0.054134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889452</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.255085</td>\n",
       "      <td>0.719518</td>\n",
       "      <td>0.445166</td>\n",
       "      <td>0.590652</td>\n",
       "      <td>0.124258</td>\n",
       "      <td>0.110540</td>\n",
       "      <td>0.822212</td>\n",
       "      <td>...</td>\n",
       "      <td>0.071035</td>\n",
       "      <td>0.911982</td>\n",
       "      <td>0.106951</td>\n",
       "      <td>0.056701</td>\n",
       "      <td>0.073805</td>\n",
       "      <td>0.063345</td>\n",
       "      <td>0.462678</td>\n",
       "      <td>0.042211</td>\n",
       "      <td>0.064198</td>\n",
       "      <td>0.061472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889453</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.916555</td>\n",
       "      <td>0.402749</td>\n",
       "      <td>0.821388</td>\n",
       "      <td>0.171016</td>\n",
       "      <td>0.819226</td>\n",
       "      <td>0.106355</td>\n",
       "      <td>0.113525</td>\n",
       "      <td>0.896862</td>\n",
       "      <td>...</td>\n",
       "      <td>0.063157</td>\n",
       "      <td>0.915611</td>\n",
       "      <td>0.103295</td>\n",
       "      <td>0.045101</td>\n",
       "      <td>0.077925</td>\n",
       "      <td>0.051017</td>\n",
       "      <td>0.421530</td>\n",
       "      <td>0.035093</td>\n",
       "      <td>0.065377</td>\n",
       "      <td>0.068653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889454</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.945348</td>\n",
       "      <td>0.269792</td>\n",
       "      <td>0.801174</td>\n",
       "      <td>0.050027</td>\n",
       "      <td>0.835881</td>\n",
       "      <td>0.115848</td>\n",
       "      <td>0.109951</td>\n",
       "      <td>0.918114</td>\n",
       "      <td>...</td>\n",
       "      <td>0.070959</td>\n",
       "      <td>0.900841</td>\n",
       "      <td>0.091191</td>\n",
       "      <td>0.043948</td>\n",
       "      <td>0.092211</td>\n",
       "      <td>0.050865</td>\n",
       "      <td>0.360305</td>\n",
       "      <td>0.039161</td>\n",
       "      <td>0.052048</td>\n",
       "      <td>0.048038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889455</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.800701</td>\n",
       "      <td>0.601968</td>\n",
       "      <td>0.671247</td>\n",
       "      <td>0.475496</td>\n",
       "      <td>0.750566</td>\n",
       "      <td>0.137646</td>\n",
       "      <td>0.123161</td>\n",
       "      <td>0.925388</td>\n",
       "      <td>...</td>\n",
       "      <td>0.062323</td>\n",
       "      <td>0.930058</td>\n",
       "      <td>0.131895</td>\n",
       "      <td>0.036567</td>\n",
       "      <td>0.071239</td>\n",
       "      <td>0.041299</td>\n",
       "      <td>0.428007</td>\n",
       "      <td>0.025063</td>\n",
       "      <td>0.096476</td>\n",
       "      <td>0.056162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889456</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.912897</td>\n",
       "      <td>0.573620</td>\n",
       "      <td>0.715421</td>\n",
       "      <td>0.424149</td>\n",
       "      <td>0.683412</td>\n",
       "      <td>0.160463</td>\n",
       "      <td>0.151035</td>\n",
       "      <td>0.896096</td>\n",
       "      <td>...</td>\n",
       "      <td>0.089045</td>\n",
       "      <td>0.864501</td>\n",
       "      <td>0.200168</td>\n",
       "      <td>0.062359</td>\n",
       "      <td>0.100639</td>\n",
       "      <td>0.065880</td>\n",
       "      <td>0.467829</td>\n",
       "      <td>0.078509</td>\n",
       "      <td>0.121617</td>\n",
       "      <td>0.086094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889457</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.905013</td>\n",
       "      <td>0.456380</td>\n",
       "      <td>0.746614</td>\n",
       "      <td>0.291481</td>\n",
       "      <td>0.700542</td>\n",
       "      <td>0.174958</td>\n",
       "      <td>0.229816</td>\n",
       "      <td>0.851584</td>\n",
       "      <td>...</td>\n",
       "      <td>0.120828</td>\n",
       "      <td>0.898401</td>\n",
       "      <td>0.199268</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>0.141701</td>\n",
       "      <td>0.060822</td>\n",
       "      <td>0.366014</td>\n",
       "      <td>0.058413</td>\n",
       "      <td>0.095742</td>\n",
       "      <td>0.077946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889458</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.945605</td>\n",
       "      <td>0.559017</td>\n",
       "      <td>0.855491</td>\n",
       "      <td>0.287494</td>\n",
       "      <td>0.789265</td>\n",
       "      <td>0.165630</td>\n",
       "      <td>0.109011</td>\n",
       "      <td>0.903462</td>\n",
       "      <td>...</td>\n",
       "      <td>0.123867</td>\n",
       "      <td>0.901532</td>\n",
       "      <td>0.178723</td>\n",
       "      <td>0.063999</td>\n",
       "      <td>0.142313</td>\n",
       "      <td>0.066441</td>\n",
       "      <td>0.445369</td>\n",
       "      <td>0.057452</td>\n",
       "      <td>0.111011</td>\n",
       "      <td>0.069320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889459</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.928470</td>\n",
       "      <td>0.190673</td>\n",
       "      <td>0.846452</td>\n",
       "      <td>0.362903</td>\n",
       "      <td>0.788598</td>\n",
       "      <td>0.151239</td>\n",
       "      <td>0.103405</td>\n",
       "      <td>0.870057</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108988</td>\n",
       "      <td>0.767391</td>\n",
       "      <td>0.126691</td>\n",
       "      <td>0.046874</td>\n",
       "      <td>0.078185</td>\n",
       "      <td>0.055519</td>\n",
       "      <td>0.416373</td>\n",
       "      <td>0.037345</td>\n",
       "      <td>0.073747</td>\n",
       "      <td>0.052308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889460</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.886674</td>\n",
       "      <td>0.472598</td>\n",
       "      <td>0.706657</td>\n",
       "      <td>0.338507</td>\n",
       "      <td>0.720732</td>\n",
       "      <td>0.141446</td>\n",
       "      <td>0.158626</td>\n",
       "      <td>0.830486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.096187</td>\n",
       "      <td>0.911952</td>\n",
       "      <td>0.143894</td>\n",
       "      <td>0.041969</td>\n",
       "      <td>0.086493</td>\n",
       "      <td>0.047331</td>\n",
       "      <td>0.402690</td>\n",
       "      <td>0.062063</td>\n",
       "      <td>0.087348</td>\n",
       "      <td>0.054760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889461</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.658577</td>\n",
       "      <td>0.428826</td>\n",
       "      <td>0.696876</td>\n",
       "      <td>0.279794</td>\n",
       "      <td>0.790845</td>\n",
       "      <td>0.135847</td>\n",
       "      <td>0.136464</td>\n",
       "      <td>0.904648</td>\n",
       "      <td>...</td>\n",
       "      <td>0.087118</td>\n",
       "      <td>0.900863</td>\n",
       "      <td>0.148636</td>\n",
       "      <td>0.044144</td>\n",
       "      <td>0.091193</td>\n",
       "      <td>0.058810</td>\n",
       "      <td>0.549300</td>\n",
       "      <td>0.033341</td>\n",
       "      <td>0.070559</td>\n",
       "      <td>0.057610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889462</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.886053</td>\n",
       "      <td>0.438095</td>\n",
       "      <td>0.809384</td>\n",
       "      <td>0.343796</td>\n",
       "      <td>0.789523</td>\n",
       "      <td>0.148287</td>\n",
       "      <td>0.146108</td>\n",
       "      <td>0.886473</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100907</td>\n",
       "      <td>0.902346</td>\n",
       "      <td>0.126748</td>\n",
       "      <td>0.043917</td>\n",
       "      <td>0.074899</td>\n",
       "      <td>0.055870</td>\n",
       "      <td>0.424691</td>\n",
       "      <td>0.036850</td>\n",
       "      <td>0.063314</td>\n",
       "      <td>0.054887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889463</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.940889</td>\n",
       "      <td>0.549648</td>\n",
       "      <td>0.828657</td>\n",
       "      <td>0.721696</td>\n",
       "      <td>0.821386</td>\n",
       "      <td>0.158198</td>\n",
       "      <td>0.170741</td>\n",
       "      <td>0.920698</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074720</td>\n",
       "      <td>0.892420</td>\n",
       "      <td>0.094953</td>\n",
       "      <td>0.039508</td>\n",
       "      <td>0.064331</td>\n",
       "      <td>0.043405</td>\n",
       "      <td>0.459248</td>\n",
       "      <td>0.033582</td>\n",
       "      <td>0.066927</td>\n",
       "      <td>0.056238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889464</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.898789</td>\n",
       "      <td>0.363438</td>\n",
       "      <td>0.843707</td>\n",
       "      <td>0.347043</td>\n",
       "      <td>0.836742</td>\n",
       "      <td>0.133636</td>\n",
       "      <td>0.212016</td>\n",
       "      <td>0.897875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.070485</td>\n",
       "      <td>0.923459</td>\n",
       "      <td>0.111249</td>\n",
       "      <td>0.040515</td>\n",
       "      <td>0.066471</td>\n",
       "      <td>0.050534</td>\n",
       "      <td>0.365591</td>\n",
       "      <td>0.035108</td>\n",
       "      <td>0.059534</td>\n",
       "      <td>0.045008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889465</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.891708</td>\n",
       "      <td>0.498447</td>\n",
       "      <td>0.752952</td>\n",
       "      <td>0.264992</td>\n",
       "      <td>0.745716</td>\n",
       "      <td>0.140901</td>\n",
       "      <td>0.134368</td>\n",
       "      <td>0.869161</td>\n",
       "      <td>...</td>\n",
       "      <td>0.068146</td>\n",
       "      <td>0.904598</td>\n",
       "      <td>0.128985</td>\n",
       "      <td>0.040090</td>\n",
       "      <td>0.080675</td>\n",
       "      <td>0.047034</td>\n",
       "      <td>0.430531</td>\n",
       "      <td>0.031282</td>\n",
       "      <td>0.092013</td>\n",
       "      <td>0.054860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889466</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.908121</td>\n",
       "      <td>0.389701</td>\n",
       "      <td>0.649820</td>\n",
       "      <td>0.237197</td>\n",
       "      <td>0.817482</td>\n",
       "      <td>0.130769</td>\n",
       "      <td>0.155981</td>\n",
       "      <td>0.908654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.080610</td>\n",
       "      <td>0.906234</td>\n",
       "      <td>0.116681</td>\n",
       "      <td>0.049301</td>\n",
       "      <td>0.076538</td>\n",
       "      <td>0.054003</td>\n",
       "      <td>0.449113</td>\n",
       "      <td>0.044718</td>\n",
       "      <td>0.075020</td>\n",
       "      <td>0.057761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889467</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.919192</td>\n",
       "      <td>0.620471</td>\n",
       "      <td>0.758308</td>\n",
       "      <td>0.416267</td>\n",
       "      <td>0.802541</td>\n",
       "      <td>0.139156</td>\n",
       "      <td>0.140242</td>\n",
       "      <td>0.918841</td>\n",
       "      <td>...</td>\n",
       "      <td>0.069710</td>\n",
       "      <td>0.764012</td>\n",
       "      <td>0.120842</td>\n",
       "      <td>0.041738</td>\n",
       "      <td>0.070390</td>\n",
       "      <td>0.041400</td>\n",
       "      <td>0.366485</td>\n",
       "      <td>0.030008</td>\n",
       "      <td>0.056330</td>\n",
       "      <td>0.053577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889468</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.884937</td>\n",
       "      <td>0.560411</td>\n",
       "      <td>0.713256</td>\n",
       "      <td>0.367997</td>\n",
       "      <td>0.728698</td>\n",
       "      <td>0.116996</td>\n",
       "      <td>0.121566</td>\n",
       "      <td>0.875237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.059863</td>\n",
       "      <td>0.879914</td>\n",
       "      <td>0.097321</td>\n",
       "      <td>0.042975</td>\n",
       "      <td>0.065390</td>\n",
       "      <td>0.043556</td>\n",
       "      <td>0.318776</td>\n",
       "      <td>0.030935</td>\n",
       "      <td>0.060639</td>\n",
       "      <td>0.046627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889469</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.847137</td>\n",
       "      <td>0.486833</td>\n",
       "      <td>0.765910</td>\n",
       "      <td>0.208881</td>\n",
       "      <td>0.747610</td>\n",
       "      <td>0.118512</td>\n",
       "      <td>0.128749</td>\n",
       "      <td>0.861853</td>\n",
       "      <td>...</td>\n",
       "      <td>0.068376</td>\n",
       "      <td>0.891084</td>\n",
       "      <td>0.120937</td>\n",
       "      <td>0.045275</td>\n",
       "      <td>0.071781</td>\n",
       "      <td>0.046470</td>\n",
       "      <td>0.460292</td>\n",
       "      <td>0.033952</td>\n",
       "      <td>0.064908</td>\n",
       "      <td>0.048883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889470</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.891285</td>\n",
       "      <td>0.534948</td>\n",
       "      <td>0.687243</td>\n",
       "      <td>0.201777</td>\n",
       "      <td>0.694031</td>\n",
       "      <td>0.131851</td>\n",
       "      <td>0.114444</td>\n",
       "      <td>0.850275</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074215</td>\n",
       "      <td>0.731445</td>\n",
       "      <td>0.110190</td>\n",
       "      <td>0.049972</td>\n",
       "      <td>0.081874</td>\n",
       "      <td>0.051141</td>\n",
       "      <td>0.391072</td>\n",
       "      <td>0.031489</td>\n",
       "      <td>0.074731</td>\n",
       "      <td>0.050397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889471</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.857592</td>\n",
       "      <td>0.418893</td>\n",
       "      <td>0.750974</td>\n",
       "      <td>0.298984</td>\n",
       "      <td>0.774942</td>\n",
       "      <td>0.119498</td>\n",
       "      <td>0.146479</td>\n",
       "      <td>0.877464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.090363</td>\n",
       "      <td>0.906157</td>\n",
       "      <td>0.129699</td>\n",
       "      <td>0.049485</td>\n",
       "      <td>0.082714</td>\n",
       "      <td>0.055286</td>\n",
       "      <td>0.427887</td>\n",
       "      <td>0.063588</td>\n",
       "      <td>0.086259</td>\n",
       "      <td>0.050523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889472</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.902668</td>\n",
       "      <td>0.494234</td>\n",
       "      <td>0.775990</td>\n",
       "      <td>0.305142</td>\n",
       "      <td>0.797583</td>\n",
       "      <td>0.151212</td>\n",
       "      <td>0.202713</td>\n",
       "      <td>0.849082</td>\n",
       "      <td>...</td>\n",
       "      <td>0.085845</td>\n",
       "      <td>0.910145</td>\n",
       "      <td>0.162196</td>\n",
       "      <td>0.051807</td>\n",
       "      <td>0.084975</td>\n",
       "      <td>0.056899</td>\n",
       "      <td>0.358213</td>\n",
       "      <td>0.042993</td>\n",
       "      <td>0.080184</td>\n",
       "      <td>0.061212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889473</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.920242</td>\n",
       "      <td>0.520000</td>\n",
       "      <td>0.740313</td>\n",
       "      <td>0.346926</td>\n",
       "      <td>0.761922</td>\n",
       "      <td>0.158237</td>\n",
       "      <td>0.180241</td>\n",
       "      <td>0.878675</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086687</td>\n",
       "      <td>0.906681</td>\n",
       "      <td>0.132830</td>\n",
       "      <td>0.055199</td>\n",
       "      <td>0.085749</td>\n",
       "      <td>0.047285</td>\n",
       "      <td>0.421398</td>\n",
       "      <td>0.035501</td>\n",
       "      <td>0.062193</td>\n",
       "      <td>0.062806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889474</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.910186</td>\n",
       "      <td>0.344111</td>\n",
       "      <td>0.764736</td>\n",
       "      <td>0.320352</td>\n",
       "      <td>0.777011</td>\n",
       "      <td>0.132540</td>\n",
       "      <td>0.115312</td>\n",
       "      <td>0.880804</td>\n",
       "      <td>...</td>\n",
       "      <td>0.073426</td>\n",
       "      <td>0.871920</td>\n",
       "      <td>0.138156</td>\n",
       "      <td>0.050589</td>\n",
       "      <td>0.074091</td>\n",
       "      <td>0.050350</td>\n",
       "      <td>0.398823</td>\n",
       "      <td>0.038157</td>\n",
       "      <td>0.095281</td>\n",
       "      <td>0.046500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889475</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.915943</td>\n",
       "      <td>0.538893</td>\n",
       "      <td>0.801592</td>\n",
       "      <td>0.292096</td>\n",
       "      <td>0.772299</td>\n",
       "      <td>0.151140</td>\n",
       "      <td>0.163227</td>\n",
       "      <td>0.878485</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084617</td>\n",
       "      <td>0.834283</td>\n",
       "      <td>0.138477</td>\n",
       "      <td>0.047229</td>\n",
       "      <td>0.082961</td>\n",
       "      <td>0.047871</td>\n",
       "      <td>0.419172</td>\n",
       "      <td>0.034143</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.049892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889476</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.846107</td>\n",
       "      <td>0.472038</td>\n",
       "      <td>0.707455</td>\n",
       "      <td>0.312392</td>\n",
       "      <td>0.737047</td>\n",
       "      <td>0.150357</td>\n",
       "      <td>0.151145</td>\n",
       "      <td>0.833751</td>\n",
       "      <td>...</td>\n",
       "      <td>0.080896</td>\n",
       "      <td>0.883883</td>\n",
       "      <td>0.182682</td>\n",
       "      <td>0.046623</td>\n",
       "      <td>0.098250</td>\n",
       "      <td>0.046916</td>\n",
       "      <td>0.332764</td>\n",
       "      <td>0.033397</td>\n",
       "      <td>0.080246</td>\n",
       "      <td>0.059738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889477</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.938361</td>\n",
       "      <td>0.435860</td>\n",
       "      <td>0.787132</td>\n",
       "      <td>0.403942</td>\n",
       "      <td>0.710624</td>\n",
       "      <td>0.138247</td>\n",
       "      <td>0.115446</td>\n",
       "      <td>0.828710</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077847</td>\n",
       "      <td>0.779722</td>\n",
       "      <td>0.152587</td>\n",
       "      <td>0.044624</td>\n",
       "      <td>0.092718</td>\n",
       "      <td>0.049732</td>\n",
       "      <td>0.384037</td>\n",
       "      <td>0.041053</td>\n",
       "      <td>0.080090</td>\n",
       "      <td>0.051687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889478</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.925077</td>\n",
       "      <td>0.384009</td>\n",
       "      <td>0.688678</td>\n",
       "      <td>0.319418</td>\n",
       "      <td>0.690327</td>\n",
       "      <td>0.133982</td>\n",
       "      <td>0.156394</td>\n",
       "      <td>0.846363</td>\n",
       "      <td>...</td>\n",
       "      <td>0.099963</td>\n",
       "      <td>0.839100</td>\n",
       "      <td>0.161876</td>\n",
       "      <td>0.047782</td>\n",
       "      <td>0.107602</td>\n",
       "      <td>0.046864</td>\n",
       "      <td>0.369536</td>\n",
       "      <td>0.037380</td>\n",
       "      <td>0.058187</td>\n",
       "      <td>0.057385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889479</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.881042</td>\n",
       "      <td>0.397866</td>\n",
       "      <td>0.743923</td>\n",
       "      <td>0.286642</td>\n",
       "      <td>0.700416</td>\n",
       "      <td>0.159335</td>\n",
       "      <td>0.150518</td>\n",
       "      <td>0.883626</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086429</td>\n",
       "      <td>0.880513</td>\n",
       "      <td>0.169100</td>\n",
       "      <td>0.043887</td>\n",
       "      <td>0.105037</td>\n",
       "      <td>0.052951</td>\n",
       "      <td>0.369721</td>\n",
       "      <td>0.038713</td>\n",
       "      <td>0.064082</td>\n",
       "      <td>0.051196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889480</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.906619</td>\n",
       "      <td>0.337584</td>\n",
       "      <td>0.814194</td>\n",
       "      <td>0.359467</td>\n",
       "      <td>0.789697</td>\n",
       "      <td>0.155050</td>\n",
       "      <td>0.155381</td>\n",
       "      <td>0.910744</td>\n",
       "      <td>...</td>\n",
       "      <td>0.085328</td>\n",
       "      <td>0.865472</td>\n",
       "      <td>0.135583</td>\n",
       "      <td>0.048151</td>\n",
       "      <td>0.079514</td>\n",
       "      <td>0.061927</td>\n",
       "      <td>0.399862</td>\n",
       "      <td>0.054132</td>\n",
       "      <td>0.144723</td>\n",
       "      <td>0.060820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889481</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.814414</td>\n",
       "      <td>0.363783</td>\n",
       "      <td>0.681041</td>\n",
       "      <td>0.342527</td>\n",
       "      <td>0.676371</td>\n",
       "      <td>0.160228</td>\n",
       "      <td>0.143462</td>\n",
       "      <td>0.843828</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142595</td>\n",
       "      <td>0.838000</td>\n",
       "      <td>0.158284</td>\n",
       "      <td>0.055035</td>\n",
       "      <td>0.090422</td>\n",
       "      <td>0.057429</td>\n",
       "      <td>0.449676</td>\n",
       "      <td>0.047198</td>\n",
       "      <td>0.094308</td>\n",
       "      <td>0.063266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889482</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.890737</td>\n",
       "      <td>0.427695</td>\n",
       "      <td>0.547426</td>\n",
       "      <td>0.306420</td>\n",
       "      <td>0.646292</td>\n",
       "      <td>0.175532</td>\n",
       "      <td>0.227513</td>\n",
       "      <td>0.865597</td>\n",
       "      <td>...</td>\n",
       "      <td>0.109913</td>\n",
       "      <td>0.883990</td>\n",
       "      <td>0.218800</td>\n",
       "      <td>0.057139</td>\n",
       "      <td>0.160496</td>\n",
       "      <td>0.058510</td>\n",
       "      <td>0.346426</td>\n",
       "      <td>0.061896</td>\n",
       "      <td>0.100295</td>\n",
       "      <td>0.084454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889483</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.775418</td>\n",
       "      <td>0.473264</td>\n",
       "      <td>0.732379</td>\n",
       "      <td>0.382156</td>\n",
       "      <td>0.768062</td>\n",
       "      <td>0.152933</td>\n",
       "      <td>0.164015</td>\n",
       "      <td>0.863262</td>\n",
       "      <td>...</td>\n",
       "      <td>0.096477</td>\n",
       "      <td>0.716387</td>\n",
       "      <td>0.172309</td>\n",
       "      <td>0.060820</td>\n",
       "      <td>0.113924</td>\n",
       "      <td>0.056668</td>\n",
       "      <td>0.405079</td>\n",
       "      <td>0.050344</td>\n",
       "      <td>0.118933</td>\n",
       "      <td>0.069341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889484</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.609464</td>\n",
       "      <td>0.414352</td>\n",
       "      <td>0.728085</td>\n",
       "      <td>0.371309</td>\n",
       "      <td>0.742481</td>\n",
       "      <td>0.161358</td>\n",
       "      <td>0.156761</td>\n",
       "      <td>0.832500</td>\n",
       "      <td>...</td>\n",
       "      <td>0.087106</td>\n",
       "      <td>0.823801</td>\n",
       "      <td>0.153061</td>\n",
       "      <td>0.052448</td>\n",
       "      <td>0.104985</td>\n",
       "      <td>0.048632</td>\n",
       "      <td>0.447421</td>\n",
       "      <td>0.037829</td>\n",
       "      <td>0.086164</td>\n",
       "      <td>0.069776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889485</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.909835</td>\n",
       "      <td>0.301115</td>\n",
       "      <td>0.755285</td>\n",
       "      <td>0.335879</td>\n",
       "      <td>0.785946</td>\n",
       "      <td>0.127912</td>\n",
       "      <td>0.122499</td>\n",
       "      <td>0.852137</td>\n",
       "      <td>...</td>\n",
       "      <td>0.071193</td>\n",
       "      <td>0.881395</td>\n",
       "      <td>0.152164</td>\n",
       "      <td>0.053912</td>\n",
       "      <td>0.088443</td>\n",
       "      <td>0.050794</td>\n",
       "      <td>0.421397</td>\n",
       "      <td>0.034190</td>\n",
       "      <td>0.100369</td>\n",
       "      <td>0.055138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889486</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.908542</td>\n",
       "      <td>0.429467</td>\n",
       "      <td>0.723343</td>\n",
       "      <td>0.314983</td>\n",
       "      <td>0.809538</td>\n",
       "      <td>0.163645</td>\n",
       "      <td>0.158484</td>\n",
       "      <td>0.880032</td>\n",
       "      <td>...</td>\n",
       "      <td>0.079157</td>\n",
       "      <td>0.883963</td>\n",
       "      <td>0.146392</td>\n",
       "      <td>0.045936</td>\n",
       "      <td>0.092081</td>\n",
       "      <td>0.049920</td>\n",
       "      <td>0.404630</td>\n",
       "      <td>0.034174</td>\n",
       "      <td>0.125893</td>\n",
       "      <td>0.054799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889487</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.770064</td>\n",
       "      <td>0.366102</td>\n",
       "      <td>0.746495</td>\n",
       "      <td>0.302081</td>\n",
       "      <td>0.741358</td>\n",
       "      <td>0.142327</td>\n",
       "      <td>0.150737</td>\n",
       "      <td>0.863217</td>\n",
       "      <td>...</td>\n",
       "      <td>0.093580</td>\n",
       "      <td>0.890962</td>\n",
       "      <td>0.152227</td>\n",
       "      <td>0.051479</td>\n",
       "      <td>0.087377</td>\n",
       "      <td>0.060809</td>\n",
       "      <td>0.384359</td>\n",
       "      <td>0.049541</td>\n",
       "      <td>0.076795</td>\n",
       "      <td>0.066625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889488</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.901580</td>\n",
       "      <td>0.558921</td>\n",
       "      <td>0.798775</td>\n",
       "      <td>0.317394</td>\n",
       "      <td>0.797128</td>\n",
       "      <td>0.141997</td>\n",
       "      <td>0.125966</td>\n",
       "      <td>0.890686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.085907</td>\n",
       "      <td>0.904276</td>\n",
       "      <td>0.106502</td>\n",
       "      <td>0.051514</td>\n",
       "      <td>0.063500</td>\n",
       "      <td>0.050605</td>\n",
       "      <td>0.463203</td>\n",
       "      <td>0.040220</td>\n",
       "      <td>0.083921</td>\n",
       "      <td>0.049213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889489</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.836649</td>\n",
       "      <td>0.368966</td>\n",
       "      <td>0.800428</td>\n",
       "      <td>0.357834</td>\n",
       "      <td>0.759670</td>\n",
       "      <td>0.136360</td>\n",
       "      <td>0.174562</td>\n",
       "      <td>0.884354</td>\n",
       "      <td>...</td>\n",
       "      <td>0.087908</td>\n",
       "      <td>0.892865</td>\n",
       "      <td>0.139733</td>\n",
       "      <td>0.050456</td>\n",
       "      <td>0.086623</td>\n",
       "      <td>0.049655</td>\n",
       "      <td>0.400643</td>\n",
       "      <td>0.030688</td>\n",
       "      <td>0.100283</td>\n",
       "      <td>0.054959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889490</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.881275</td>\n",
       "      <td>0.398305</td>\n",
       "      <td>0.767000</td>\n",
       "      <td>0.334070</td>\n",
       "      <td>0.774546</td>\n",
       "      <td>0.148788</td>\n",
       "      <td>0.139080</td>\n",
       "      <td>0.874481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.065351</td>\n",
       "      <td>0.877259</td>\n",
       "      <td>0.115265</td>\n",
       "      <td>0.044668</td>\n",
       "      <td>0.078492</td>\n",
       "      <td>0.052283</td>\n",
       "      <td>0.387745</td>\n",
       "      <td>0.031048</td>\n",
       "      <td>0.091715</td>\n",
       "      <td>0.047587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM889491</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.840217</td>\n",
       "      <td>0.449060</td>\n",
       "      <td>0.759809</td>\n",
       "      <td>0.320748</td>\n",
       "      <td>0.740727</td>\n",
       "      <td>0.131508</td>\n",
       "      <td>0.147820</td>\n",
       "      <td>0.899412</td>\n",
       "      <td>...</td>\n",
       "      <td>0.071807</td>\n",
       "      <td>0.893780</td>\n",
       "      <td>0.147599</td>\n",
       "      <td>0.043283</td>\n",
       "      <td>0.097718</td>\n",
       "      <td>0.048284</td>\n",
       "      <td>0.397563</td>\n",
       "      <td>0.036458</td>\n",
       "      <td>0.083698</td>\n",
       "      <td>0.055029</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48 rows Ã— 11146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer  Gender      A1BG       A2M     A2ML1  \\\n",
       "GSM889444                    1.0     1.0  0.877513  0.488576  0.770778   \n",
       "GSM889445                    1.0     1.0  0.790216  0.616612  0.822486   \n",
       "GSM889446                    1.0     1.0  0.905854  0.415186  0.758853   \n",
       "GSM889447                    1.0     1.0  0.926305  0.450625  0.844311   \n",
       "GSM889448                    1.0     1.0  0.813567  0.524828  0.796473   \n",
       "GSM889449                    1.0     0.0  0.935604  0.255921  0.622104   \n",
       "GSM889450                    1.0     1.0  0.876431  0.446866  0.740103   \n",
       "GSM889451                    1.0     0.0  0.900982  0.419583  0.874109   \n",
       "GSM889452                    1.0     0.0  0.923077  0.255085  0.719518   \n",
       "GSM889453                    1.0     1.0  0.916555  0.402749  0.821388   \n",
       "GSM889454                    1.0     1.0  0.945348  0.269792  0.801174   \n",
       "GSM889455                    1.0     1.0  0.800701  0.601968  0.671247   \n",
       "GSM889456                    1.0     1.0  0.912897  0.573620  0.715421   \n",
       "GSM889457                    1.0     0.0  0.905013  0.456380  0.746614   \n",
       "GSM889458                    1.0     0.0  0.945605  0.559017  0.855491   \n",
       "GSM889459                    1.0     1.0  0.928470  0.190673  0.846452   \n",
       "GSM889460                    1.0     1.0  0.886674  0.472598  0.706657   \n",
       "GSM889461                    1.0     1.0  0.658577  0.428826  0.696876   \n",
       "GSM889462                    1.0     1.0  0.886053  0.438095  0.809384   \n",
       "GSM889463                    1.0     0.0  0.940889  0.549648  0.828657   \n",
       "GSM889464                    1.0     0.0  0.898789  0.363438  0.843707   \n",
       "GSM889465                    1.0     1.0  0.891708  0.498447  0.752952   \n",
       "GSM889466                    1.0     1.0  0.908121  0.389701  0.649820   \n",
       "GSM889467                    1.0     1.0  0.919192  0.620471  0.758308   \n",
       "GSM889468                    1.0     0.0  0.884937  0.560411  0.713256   \n",
       "GSM889469                    1.0     1.0  0.847137  0.486833  0.765910   \n",
       "GSM889470                    1.0     0.0  0.891285  0.534948  0.687243   \n",
       "GSM889471                    1.0     1.0  0.857592  0.418893  0.750974   \n",
       "GSM889472                    1.0     1.0  0.902668  0.494234  0.775990   \n",
       "GSM889473                    1.0     0.0  0.920242  0.520000  0.740313   \n",
       "GSM889474                    1.0     1.0  0.910186  0.344111  0.764736   \n",
       "GSM889475                    1.0     1.0  0.915943  0.538893  0.801592   \n",
       "GSM889476                    1.0     1.0  0.846107  0.472038  0.707455   \n",
       "GSM889477                    1.0     1.0  0.938361  0.435860  0.787132   \n",
       "GSM889478                    1.0     0.0  0.925077  0.384009  0.688678   \n",
       "GSM889479                    1.0     0.0  0.881042  0.397866  0.743923   \n",
       "GSM889480                    1.0     1.0  0.906619  0.337584  0.814194   \n",
       "GSM889481                    1.0     1.0  0.814414  0.363783  0.681041   \n",
       "GSM889482                    1.0     1.0  0.890737  0.427695  0.547426   \n",
       "GSM889483                    1.0     1.0  0.775418  0.473264  0.732379   \n",
       "GSM889484                    1.0     1.0  0.609464  0.414352  0.728085   \n",
       "GSM889485                    1.0     0.0  0.909835  0.301115  0.755285   \n",
       "GSM889486                    0.0     0.0  0.908542  0.429467  0.723343   \n",
       "GSM889487                    0.0     1.0  0.770064  0.366102  0.746495   \n",
       "GSM889488                    0.0     0.0  0.901580  0.558921  0.798775   \n",
       "GSM889489                    0.0     1.0  0.836649  0.368966  0.800428   \n",
       "GSM889490                    0.0     0.0  0.881275  0.398305  0.767000   \n",
       "GSM889491                    0.0     1.0  0.840217  0.449060  0.759809   \n",
       "\n",
       "             A4GALT     A4GNT      AAAS      AACS     AADAC  ...    ZSCAN2  \\\n",
       "GSM889444  0.337080  0.731916  0.152761  0.145945  0.883583  ...  0.060651   \n",
       "GSM889445  0.623585  0.760295  0.155780  0.222403  0.916509  ...  0.078385   \n",
       "GSM889446  0.400251  0.797876  0.135304  0.105125  0.853768  ...  0.090061   \n",
       "GSM889447  0.422990  0.610334  0.123305  0.076724  0.693930  ...  0.061925   \n",
       "GSM889448  0.342717  0.804778  0.091906  0.106987  0.864323  ...  0.072116   \n",
       "GSM889449  0.503112  0.720754  0.104102  0.098242  0.771726  ...  0.051829   \n",
       "GSM889450  0.257814  0.722959  0.163676  0.144049  0.902074  ...  0.101449   \n",
       "GSM889451  0.359270  0.862955  0.117089  0.104022  0.932222  ...  0.076846   \n",
       "GSM889452  0.445166  0.590652  0.124258  0.110540  0.822212  ...  0.071035   \n",
       "GSM889453  0.171016  0.819226  0.106355  0.113525  0.896862  ...  0.063157   \n",
       "GSM889454  0.050027  0.835881  0.115848  0.109951  0.918114  ...  0.070959   \n",
       "GSM889455  0.475496  0.750566  0.137646  0.123161  0.925388  ...  0.062323   \n",
       "GSM889456  0.424149  0.683412  0.160463  0.151035  0.896096  ...  0.089045   \n",
       "GSM889457  0.291481  0.700542  0.174958  0.229816  0.851584  ...  0.120828   \n",
       "GSM889458  0.287494  0.789265  0.165630  0.109011  0.903462  ...  0.123867   \n",
       "GSM889459  0.362903  0.788598  0.151239  0.103405  0.870057  ...  0.108988   \n",
       "GSM889460  0.338507  0.720732  0.141446  0.158626  0.830486  ...  0.096187   \n",
       "GSM889461  0.279794  0.790845  0.135847  0.136464  0.904648  ...  0.087118   \n",
       "GSM889462  0.343796  0.789523  0.148287  0.146108  0.886473  ...  0.100907   \n",
       "GSM889463  0.721696  0.821386  0.158198  0.170741  0.920698  ...  0.074720   \n",
       "GSM889464  0.347043  0.836742  0.133636  0.212016  0.897875  ...  0.070485   \n",
       "GSM889465  0.264992  0.745716  0.140901  0.134368  0.869161  ...  0.068146   \n",
       "GSM889466  0.237197  0.817482  0.130769  0.155981  0.908654  ...  0.080610   \n",
       "GSM889467  0.416267  0.802541  0.139156  0.140242  0.918841  ...  0.069710   \n",
       "GSM889468  0.367997  0.728698  0.116996  0.121566  0.875237  ...  0.059863   \n",
       "GSM889469  0.208881  0.747610  0.118512  0.128749  0.861853  ...  0.068376   \n",
       "GSM889470  0.201777  0.694031  0.131851  0.114444  0.850275  ...  0.074215   \n",
       "GSM889471  0.298984  0.774942  0.119498  0.146479  0.877464  ...  0.090363   \n",
       "GSM889472  0.305142  0.797583  0.151212  0.202713  0.849082  ...  0.085845   \n",
       "GSM889473  0.346926  0.761922  0.158237  0.180241  0.878675  ...  0.086687   \n",
       "GSM889474  0.320352  0.777011  0.132540  0.115312  0.880804  ...  0.073426   \n",
       "GSM889475  0.292096  0.772299  0.151140  0.163227  0.878485  ...  0.084617   \n",
       "GSM889476  0.312392  0.737047  0.150357  0.151145  0.833751  ...  0.080896   \n",
       "GSM889477  0.403942  0.710624  0.138247  0.115446  0.828710  ...  0.077847   \n",
       "GSM889478  0.319418  0.690327  0.133982  0.156394  0.846363  ...  0.099963   \n",
       "GSM889479  0.286642  0.700416  0.159335  0.150518  0.883626  ...  0.086429   \n",
       "GSM889480  0.359467  0.789697  0.155050  0.155381  0.910744  ...  0.085328   \n",
       "GSM889481  0.342527  0.676371  0.160228  0.143462  0.843828  ...  0.142595   \n",
       "GSM889482  0.306420  0.646292  0.175532  0.227513  0.865597  ...  0.109913   \n",
       "GSM889483  0.382156  0.768062  0.152933  0.164015  0.863262  ...  0.096477   \n",
       "GSM889484  0.371309  0.742481  0.161358  0.156761  0.832500  ...  0.087106   \n",
       "GSM889485  0.335879  0.785946  0.127912  0.122499  0.852137  ...  0.071193   \n",
       "GSM889486  0.314983  0.809538  0.163645  0.158484  0.880032  ...  0.079157   \n",
       "GSM889487  0.302081  0.741358  0.142327  0.150737  0.863217  ...  0.093580   \n",
       "GSM889488  0.317394  0.797128  0.141997  0.125966  0.890686  ...  0.085907   \n",
       "GSM889489  0.357834  0.759670  0.136360  0.174562  0.884354  ...  0.087908   \n",
       "GSM889490  0.334070  0.774546  0.148788  0.139080  0.874481  ...  0.065351   \n",
       "GSM889491  0.320748  0.740727  0.131508  0.147820  0.899412  ...  0.071807   \n",
       "\n",
       "             ZSCAN4    ZSWIM1      ZW10    ZWILCH     ZWINT      ZXDA  \\\n",
       "GSM889444  0.861556  0.113921  0.039714  0.068936  0.055689  0.471868   \n",
       "GSM889445  0.597086  0.099773  0.044380  0.077357  0.054493  0.394791   \n",
       "GSM889446  0.935782  0.100392  0.050342  0.099744  0.046531  0.417707   \n",
       "GSM889447  0.113502  0.082969  0.042907  0.059283  0.050910  0.432613   \n",
       "GSM889448  0.926427  0.096655  0.049146  0.077697  0.045079  0.440079   \n",
       "GSM889449  0.922562  0.070781  0.041381  0.067504  0.039365  0.430349   \n",
       "GSM889450  0.865154  0.165601  0.050697  0.126225  0.099023  0.407978   \n",
       "GSM889451  0.939495  0.100747  0.043479  0.089807  0.051996  0.394574   \n",
       "GSM889452  0.911982  0.106951  0.056701  0.073805  0.063345  0.462678   \n",
       "GSM889453  0.915611  0.103295  0.045101  0.077925  0.051017  0.421530   \n",
       "GSM889454  0.900841  0.091191  0.043948  0.092211  0.050865  0.360305   \n",
       "GSM889455  0.930058  0.131895  0.036567  0.071239  0.041299  0.428007   \n",
       "GSM889456  0.864501  0.200168  0.062359  0.100639  0.065880  0.467829   \n",
       "GSM889457  0.898401  0.199268  0.071429  0.141701  0.060822  0.366014   \n",
       "GSM889458  0.901532  0.178723  0.063999  0.142313  0.066441  0.445369   \n",
       "GSM889459  0.767391  0.126691  0.046874  0.078185  0.055519  0.416373   \n",
       "GSM889460  0.911952  0.143894  0.041969  0.086493  0.047331  0.402690   \n",
       "GSM889461  0.900863  0.148636  0.044144  0.091193  0.058810  0.549300   \n",
       "GSM889462  0.902346  0.126748  0.043917  0.074899  0.055870  0.424691   \n",
       "GSM889463  0.892420  0.094953  0.039508  0.064331  0.043405  0.459248   \n",
       "GSM889464  0.923459  0.111249  0.040515  0.066471  0.050534  0.365591   \n",
       "GSM889465  0.904598  0.128985  0.040090  0.080675  0.047034  0.430531   \n",
       "GSM889466  0.906234  0.116681  0.049301  0.076538  0.054003  0.449113   \n",
       "GSM889467  0.764012  0.120842  0.041738  0.070390  0.041400  0.366485   \n",
       "GSM889468  0.879914  0.097321  0.042975  0.065390  0.043556  0.318776   \n",
       "GSM889469  0.891084  0.120937  0.045275  0.071781  0.046470  0.460292   \n",
       "GSM889470  0.731445  0.110190  0.049972  0.081874  0.051141  0.391072   \n",
       "GSM889471  0.906157  0.129699  0.049485  0.082714  0.055286  0.427887   \n",
       "GSM889472  0.910145  0.162196  0.051807  0.084975  0.056899  0.358213   \n",
       "GSM889473  0.906681  0.132830  0.055199  0.085749  0.047285  0.421398   \n",
       "GSM889474  0.871920  0.138156  0.050589  0.074091  0.050350  0.398823   \n",
       "GSM889475  0.834283  0.138477  0.047229  0.082961  0.047871  0.419172   \n",
       "GSM889476  0.883883  0.182682  0.046623  0.098250  0.046916  0.332764   \n",
       "GSM889477  0.779722  0.152587  0.044624  0.092718  0.049732  0.384037   \n",
       "GSM889478  0.839100  0.161876  0.047782  0.107602  0.046864  0.369536   \n",
       "GSM889479  0.880513  0.169100  0.043887  0.105037  0.052951  0.369721   \n",
       "GSM889480  0.865472  0.135583  0.048151  0.079514  0.061927  0.399862   \n",
       "GSM889481  0.838000  0.158284  0.055035  0.090422  0.057429  0.449676   \n",
       "GSM889482  0.883990  0.218800  0.057139  0.160496  0.058510  0.346426   \n",
       "GSM889483  0.716387  0.172309  0.060820  0.113924  0.056668  0.405079   \n",
       "GSM889484  0.823801  0.153061  0.052448  0.104985  0.048632  0.447421   \n",
       "GSM889485  0.881395  0.152164  0.053912  0.088443  0.050794  0.421397   \n",
       "GSM889486  0.883963  0.146392  0.045936  0.092081  0.049920  0.404630   \n",
       "GSM889487  0.890962  0.152227  0.051479  0.087377  0.060809  0.384359   \n",
       "GSM889488  0.904276  0.106502  0.051514  0.063500  0.050605  0.463203   \n",
       "GSM889489  0.892865  0.139733  0.050456  0.086623  0.049655  0.400643   \n",
       "GSM889490  0.877259  0.115265  0.044668  0.078492  0.052283  0.387745   \n",
       "GSM889491  0.893780  0.147599  0.043283  0.097718  0.048284  0.397563   \n",
       "\n",
       "                ZYX     ZZEF1      ZZZ3  \n",
       "GSM889444  0.028720  0.080601  0.050376  \n",
       "GSM889445  0.037781  0.086067  0.049650  \n",
       "GSM889446  0.039477  0.051746  0.061866  \n",
       "GSM889447  0.034675  0.062786  0.061720  \n",
       "GSM889448  0.032816  0.406175  0.057205  \n",
       "GSM889449  0.038400  0.085561  0.049556  \n",
       "GSM889450  0.133074  0.072221  0.104789  \n",
       "GSM889451  0.034445  0.066332  0.054134  \n",
       "GSM889452  0.042211  0.064198  0.061472  \n",
       "GSM889453  0.035093  0.065377  0.068653  \n",
       "GSM889454  0.039161  0.052048  0.048038  \n",
       "GSM889455  0.025063  0.096476  0.056162  \n",
       "GSM889456  0.078509  0.121617  0.086094  \n",
       "GSM889457  0.058413  0.095742  0.077946  \n",
       "GSM889458  0.057452  0.111011  0.069320  \n",
       "GSM889459  0.037345  0.073747  0.052308  \n",
       "GSM889460  0.062063  0.087348  0.054760  \n",
       "GSM889461  0.033341  0.070559  0.057610  \n",
       "GSM889462  0.036850  0.063314  0.054887  \n",
       "GSM889463  0.033582  0.066927  0.056238  \n",
       "GSM889464  0.035108  0.059534  0.045008  \n",
       "GSM889465  0.031282  0.092013  0.054860  \n",
       "GSM889466  0.044718  0.075020  0.057761  \n",
       "GSM889467  0.030008  0.056330  0.053577  \n",
       "GSM889468  0.030935  0.060639  0.046627  \n",
       "GSM889469  0.033952  0.064908  0.048883  \n",
       "GSM889470  0.031489  0.074731  0.050397  \n",
       "GSM889471  0.063588  0.086259  0.050523  \n",
       "GSM889472  0.042993  0.080184  0.061212  \n",
       "GSM889473  0.035501  0.062193  0.062806  \n",
       "GSM889474  0.038157  0.095281  0.046500  \n",
       "GSM889475  0.034143  0.067823  0.049892  \n",
       "GSM889476  0.033397  0.080246  0.059738  \n",
       "GSM889477  0.041053  0.080090  0.051687  \n",
       "GSM889478  0.037380  0.058187  0.057385  \n",
       "GSM889479  0.038713  0.064082  0.051196  \n",
       "GSM889480  0.054132  0.144723  0.060820  \n",
       "GSM889481  0.047198  0.094308  0.063266  \n",
       "GSM889482  0.061896  0.100295  0.084454  \n",
       "GSM889483  0.050344  0.118933  0.069341  \n",
       "GSM889484  0.037829  0.086164  0.069776  \n",
       "GSM889485  0.034190  0.100369  0.055138  \n",
       "GSM889486  0.034174  0.125893  0.054799  \n",
       "GSM889487  0.049541  0.076795  0.066625  \n",
       "GSM889488  0.040220  0.083921  0.049213  \n",
       "GSM889489  0.030688  0.100283  0.054959  \n",
       "GSM889490  0.031048  0.091715  0.047587  \n",
       "GSM889491  0.036458  0.083698  0.055029  \n",
       "\n",
       "[48 rows x 11146 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 48 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '0.0' with 6 occurrences. This represents 12.50% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 16 occurrences. This represents 33.33% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Adrenocortical Carcinoma Gene Expression Profiling\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: 2</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 2</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>...</td>\n",
       "      <td>Stage: 1</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 2</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: 3</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor grade: 1</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 2</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: Cushings</td>\n",
       "      <td>functional: Cushings</td>\n",
       "      <td>...</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: Unknown</td>\n",
       "      <td>functional: Cortisol, aldosterone, testosterone</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: aldosterone</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: Unknown</td>\n",
       "      <td>functional: Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: 23.3</td>\n",
       "      <td>age in years: 56.5</td>\n",
       "      <td>age in years: 67.8</td>\n",
       "      <td>age in years: 72.1</td>\n",
       "      <td>age in years: 46.9</td>\n",
       "      <td>...</td>\n",
       "      <td>age in years: 57</td>\n",
       "      <td>age in years: 59</td>\n",
       "      <td>age in years: 59</td>\n",
       "      <td>age in years: 55</td>\n",
       "      <td>age in years: 51</td>\n",
       "      <td>age in years: 53</td>\n",
       "      <td>age in years: 69</td>\n",
       "      <td>age in years: 63</td>\n",
       "      <td>age in years: 28</td>\n",
       "      <td>age in years: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: 3</td>\n",
       "      <td>survival in years: 0.6</td>\n",
       "      <td>survival in years: 1.7</td>\n",
       "      <td>survival in years: 0.4</td>\n",
       "      <td>survival in years: 0.1</td>\n",
       "      <td>...</td>\n",
       "      <td>survival in years: 3</td>\n",
       "      <td>survival in years: 7.583</td>\n",
       "      <td>survival in years: Unknown</td>\n",
       "      <td>survival in years: 0.583</td>\n",
       "      <td>survival in years: 6</td>\n",
       "      <td>survival in years: 2.083</td>\n",
       "      <td>survival in years: 2.83</td>\n",
       "      <td>survival in years: 2.08</td>\n",
       "      <td>survival in years: Unknown</td>\n",
       "      <td>survival in years: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>...</td>\n",
       "      <td>survival status: alive</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: Unknown</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: alive</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: alive</td>\n",
       "      <td>survival status: Unknown</td>\n",
       "      <td>survival status: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: 19</td>\n",
       "      <td>tumor size in cm: 9</td>\n",
       "      <td>tumor size in cm: 7.6</td>\n",
       "      <td>tumor size in cm: 9.5</td>\n",
       "      <td>tumor size in cm: 12</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor size in cm: 4</td>\n",
       "      <td>tumor size in cm: 2.5</td>\n",
       "      <td>tumor size in cm: 10</td>\n",
       "      <td>tumor size in cm: 10.5</td>\n",
       "      <td>tumor size in cm: 14.5</td>\n",
       "      <td>tumor size in cm: 14.5</td>\n",
       "      <td>tumor size in cm: 7.8</td>\n",
       "      <td>tumor size in cm: 7.8</td>\n",
       "      <td>tumor size in cm: Unknown</td>\n",
       "      <td>tumor size in cm: Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: 1100</td>\n",
       "      <td>tumor weight in grams: 190</td>\n",
       "      <td>tumor weight in grams: 150</td>\n",
       "      <td>tumor weight in grams: 175</td>\n",
       "      <td>tumor weight in grams: 235</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor weight in grams: 39</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "      <td>tumor weight in grams: 22</td>\n",
       "      <td>tumor weight in grams: 277</td>\n",
       "      <td>tumor weight in grams: 325</td>\n",
       "      <td>tumor weight in grams: 1243</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "      <td>tumor weight in grams: 132</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                  GSM493251  \\\n",
       "0  !Sample_characteristics_ch1                  Stage: NA   \n",
       "1  !Sample_characteristics_ch1            tumor grade: NA   \n",
       "2  !Sample_characteristics_ch1             functional: NA   \n",
       "3  !Sample_characteristics_ch1            gender: Unknown   \n",
       "4  !Sample_characteristics_ch1      age in years: Unknown   \n",
       "5  !Sample_characteristics_ch1      survival in years: NA   \n",
       "6  !Sample_characteristics_ch1        survival status: NA   \n",
       "7  !Sample_characteristics_ch1       tumor size in cm: NA   \n",
       "8  !Sample_characteristics_ch1  tumor weight in grams: NA   \n",
       "9  !Sample_characteristics_ch1                   batch: 1   \n",
       "\n",
       "                   GSM493252                  GSM493253  \\\n",
       "0                  Stage: NA                  Stage: NA   \n",
       "1            tumor grade: NA            tumor grade: NA   \n",
       "2             functional: NA             functional: NA   \n",
       "3            gender: Unknown            gender: Unknown   \n",
       "4      age in years: Unknown      age in years: Unknown   \n",
       "5      survival in years: NA      survival in years: NA   \n",
       "6        survival status: NA        survival status: NA   \n",
       "7       tumor size in cm: NA       tumor size in cm: NA   \n",
       "8  tumor weight in grams: NA  tumor weight in grams: NA   \n",
       "9                   batch: 1                   batch: 1   \n",
       "\n",
       "                   GSM493254                    GSM493255  \\\n",
       "0                  Stage: NA                     Stage: 2   \n",
       "1            tumor grade: NA               tumor grade: 3   \n",
       "2             functional: NA             functional: None   \n",
       "3            gender: Unknown                    gender: M   \n",
       "4      age in years: Unknown           age in years: 23.3   \n",
       "5      survival in years: NA         survival in years: 3   \n",
       "6        survival status: NA        survival status: dead   \n",
       "7       tumor size in cm: NA         tumor size in cm: 19   \n",
       "8  tumor weight in grams: NA  tumor weight in grams: 1100   \n",
       "9                   batch: 1                     batch: 1   \n",
       "\n",
       "                    GSM493256                   GSM493257  \\\n",
       "0                    Stage: 4                    Stage: 2   \n",
       "1              tumor grade: 4              tumor grade: 4   \n",
       "2            functional: None            functional: None   \n",
       "3                   gender: F                   gender: M   \n",
       "4          age in years: 56.5          age in years: 67.8   \n",
       "5      survival in years: 0.6      survival in years: 1.7   \n",
       "6       survival status: dead       survival status: dead   \n",
       "7         tumor size in cm: 9       tumor size in cm: 7.6   \n",
       "8  tumor weight in grams: 190  tumor weight in grams: 150   \n",
       "9                    batch: 1                    batch: 1   \n",
       "\n",
       "                    GSM493258                   GSM493259  ...  \\\n",
       "0                    Stage: 4                    Stage: 4  ...   \n",
       "1              tumor grade: 4              tumor grade: 4  ...   \n",
       "2        functional: Cushings        functional: Cushings  ...   \n",
       "3                   gender: M                   gender: M  ...   \n",
       "4          age in years: 72.1          age in years: 46.9  ...   \n",
       "5      survival in years: 0.4      survival in years: 0.1  ...   \n",
       "6       survival status: dead       survival status: dead  ...   \n",
       "7       tumor size in cm: 9.5        tumor size in cm: 12  ...   \n",
       "8  tumor weight in grams: 175  tumor weight in grams: 235  ...   \n",
       "9                    batch: 1                    batch: 1  ...   \n",
       "\n",
       "                  GSM1094071                      GSM1094072  \\\n",
       "0                   Stage: 1               Stage: Recurrence   \n",
       "1             tumor grade: 1            tumor grade: Unknown   \n",
       "2           functional: None                functional: None   \n",
       "3                  gender: M                       gender: M   \n",
       "4           age in years: 57                age in years: 59   \n",
       "5       survival in years: 3        survival in years: 7.583   \n",
       "6     survival status: alive           survival status: dead   \n",
       "7        tumor size in cm: 4           tumor size in cm: 2.5   \n",
       "8  tumor weight in grams: 39  tumor weight in grams: unknown   \n",
       "9                   batch: 2                        batch: 2   \n",
       "\n",
       "                   GSM1094073  \\\n",
       "0                    Stage: 4   \n",
       "1        tumor grade: Unknown   \n",
       "2         functional: Unknown   \n",
       "3                   gender: F   \n",
       "4            age in years: 59   \n",
       "5  survival in years: Unknown   \n",
       "6    survival status: Unknown   \n",
       "7        tumor size in cm: 10   \n",
       "8   tumor weight in grams: 22   \n",
       "9                    batch: 2   \n",
       "\n",
       "                                        GSM1094074  \\\n",
       "0                                         Stage: 2   \n",
       "1                                   tumor grade: 4   \n",
       "2  functional: Cortisol, aldosterone, testosterone   \n",
       "3                                        gender: F   \n",
       "4                                 age in years: 55   \n",
       "5                         survival in years: 0.583   \n",
       "6                            survival status: dead   \n",
       "7                           tumor size in cm: 10.5   \n",
       "8                       tumor weight in grams: 277   \n",
       "9                                         batch: 2   \n",
       "\n",
       "                   GSM1094075                   GSM1094076  \\\n",
       "0           Stage: Recurrence            Stage: Recurrence   \n",
       "1              tumor grade: 2         tumor grade: Unknown   \n",
       "2            functional: None      functional: aldosterone   \n",
       "3                   gender: F                    gender: F   \n",
       "4            age in years: 51             age in years: 53   \n",
       "5        survival in years: 6     survival in years: 2.083   \n",
       "6      survival status: alive        survival status: dead   \n",
       "7      tumor size in cm: 14.5       tumor size in cm: 14.5   \n",
       "8  tumor weight in grams: 325  tumor weight in grams: 1243   \n",
       "9                    batch: 2                     batch: 2   \n",
       "\n",
       "                       GSM1094077                  GSM1094078  \\\n",
       "0               Stage: Recurrence                    Stage: 4   \n",
       "1            tumor grade: Unknown              tumor grade: 4   \n",
       "2                functional: None            functional: None   \n",
       "3                       gender: M                   gender: M   \n",
       "4                age in years: 69            age in years: 63   \n",
       "5         survival in years: 2.83     survival in years: 2.08   \n",
       "6           survival status: dead      survival status: alive   \n",
       "7           tumor size in cm: 7.8       tumor size in cm: 7.8   \n",
       "8  tumor weight in grams: unknown  tumor weight in grams: 132   \n",
       "9                        batch: 2                    batch: 2   \n",
       "\n",
       "                       GSM1094079                      GSM1094080  \n",
       "0                        Stage: 4                       Stage: NA  \n",
       "1            tumor grade: Unknown            tumor grade: Unknown  \n",
       "2             functional: Unknown             functional: Unknown  \n",
       "3                       gender: F                      gender: NA  \n",
       "4                age in years: 28                age in years: NA  \n",
       "5      survival in years: Unknown           survival in years: NA  \n",
       "6        survival status: Unknown             survival status: NA  \n",
       "7       tumor size in cm: Unknown       tumor size in cm: Unknown  \n",
       "8  tumor weight in grams: unknown  tumor weight in grams: unknown  \n",
       "9                        batch: 2                        batch: 2  \n",
       "\n",
       "[10 rows x 49 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished: asked questions\n",
    "\n",
    "cohort = accession_num = \"GSE19776\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'tumor grade: NA', 'tumor grade: 3',\n",
       "       'tumor grade: 4', 'tumor grade: 2', 'tumor grade: 1',\n",
       "       'tumor grade: Unknown'], dtype=object)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 4\n",
    "gender_row = 3\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "\n",
    "def convert_trait(tumor_grade):\n",
    "    if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "        return 1  \n",
    "    elif tumor_grade == 'tumor grade: 1':\n",
    "        return 0  \n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>57</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>55</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "      <td>69</td>\n",
       "      <td>63</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      GSM493251 GSM493252 GSM493253 GSM493254 GSM493255  \\\n",
       "Adrenocortical Cancer      None      None      None      None         1   \n",
       "Age                        None      None      None      None      None   \n",
       "Gender                     None      None      None      None         0   \n",
       "\n",
       "                      GSM493256 GSM493257 GSM493258 GSM493259 GSM493260  ...  \\\n",
       "Adrenocortical Cancer         1         1         1         1         1  ...   \n",
       "Age                        None      None      None      None      None  ...   \n",
       "Gender                        1         0         0         0         1  ...   \n",
       "\n",
       "                      GSM1094071 GSM1094072 GSM1094073 GSM1094074  GSM1094075  \\\n",
       "Adrenocortical Cancer          0       None       None          1           1   \n",
       "Age                           57         59         59         55          51   \n",
       "Gender                         0          0          1          1           1   \n",
       "\n",
       "                      GSM1094076 GSM1094077 GSM1094078 GSM1094079  GSM1094080  \n",
       "Adrenocortical Cancer       None       None          1       None        None  \n",
       "Age                           53         69         63         28        None  \n",
       "Gender                         1          0          0          1        None  \n",
       "\n",
       "[3 rows x 48 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>101.10</td>\n",
       "      <td>48.86</td>\n",
       "      <td>100.83</td>\n",
       "      <td>104.84</td>\n",
       "      <td>566.78</td>\n",
       "      <td>418.37</td>\n",
       "      <td>499.31</td>\n",
       "      <td>432.83</td>\n",
       "      <td>212.42</td>\n",
       "      <td>336.85</td>\n",
       "      <td>...</td>\n",
       "      <td>139.79</td>\n",
       "      <td>35.72</td>\n",
       "      <td>79.18</td>\n",
       "      <td>154.74</td>\n",
       "      <td>166.63</td>\n",
       "      <td>29.67</td>\n",
       "      <td>72.48</td>\n",
       "      <td>10.21</td>\n",
       "      <td>151.86</td>\n",
       "      <td>137.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>22.58</td>\n",
       "      <td>18.30</td>\n",
       "      <td>16.96</td>\n",
       "      <td>16.96</td>\n",
       "      <td>18.64</td>\n",
       "      <td>39.96</td>\n",
       "      <td>57.40</td>\n",
       "      <td>34.15</td>\n",
       "      <td>43.71</td>\n",
       "      <td>32.52</td>\n",
       "      <td>...</td>\n",
       "      <td>8.32</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.64</td>\n",
       "      <td>17.01</td>\n",
       "      <td>17.19</td>\n",
       "      <td>25.84</td>\n",
       "      <td>16.13</td>\n",
       "      <td>43.64</td>\n",
       "      <td>11.46</td>\n",
       "      <td>22.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>73.33</td>\n",
       "      <td>30.19</td>\n",
       "      <td>155.69</td>\n",
       "      <td>173.28</td>\n",
       "      <td>14.43</td>\n",
       "      <td>224.29</td>\n",
       "      <td>30.10</td>\n",
       "      <td>9.52</td>\n",
       "      <td>9.17</td>\n",
       "      <td>9.09</td>\n",
       "      <td>...</td>\n",
       "      <td>24.91</td>\n",
       "      <td>11.09</td>\n",
       "      <td>309.80</td>\n",
       "      <td>42.61</td>\n",
       "      <td>9.57</td>\n",
       "      <td>38.21</td>\n",
       "      <td>50.33</td>\n",
       "      <td>321.67</td>\n",
       "      <td>14.87</td>\n",
       "      <td>9.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>11.97</td>\n",
       "      <td>10.74</td>\n",
       "      <td>9.91</td>\n",
       "      <td>9.77</td>\n",
       "      <td>8.73</td>\n",
       "      <td>13.49</td>\n",
       "      <td>8.65</td>\n",
       "      <td>9.59</td>\n",
       "      <td>12.71</td>\n",
       "      <td>8.71</td>\n",
       "      <td>...</td>\n",
       "      <td>8.73</td>\n",
       "      <td>18.09</td>\n",
       "      <td>9.59</td>\n",
       "      <td>8.69</td>\n",
       "      <td>8.69</td>\n",
       "      <td>12.21</td>\n",
       "      <td>8.64</td>\n",
       "      <td>8.64</td>\n",
       "      <td>8.68</td>\n",
       "      <td>8.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.76</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM493251  GSM493252  GSM493253  GSM493254  GSM493255  \\\n",
       "ID                                                                       \n",
       "1007_s_at           101.10      48.86     100.83     104.84     566.78   \n",
       "1053_at              22.58      18.30      16.96      16.96      18.64   \n",
       "117_at               73.33      30.19     155.69     173.28      14.43   \n",
       "121_at               11.97      10.74       9.91       9.77       8.73   \n",
       "1255_g_at             5.50       5.50       5.50       5.50       5.50   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "AFFX-ThrX-5_at        5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-ThrX-M_at        5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-TrpnX-3_at       5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-TrpnX-5_at       5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-TrpnX-M_at       5.50       5.50       5.50       5.76       5.50   \n",
       "\n",
       "                 GSM493256  GSM493257  GSM493258  GSM493259  GSM493260  ...  \\\n",
       "ID                                                                      ...   \n",
       "1007_s_at           418.37     499.31     432.83     212.42     336.85  ...   \n",
       "1053_at              39.96      57.40      34.15      43.71      32.52  ...   \n",
       "117_at              224.29      30.10       9.52       9.17       9.09  ...   \n",
       "121_at               13.49       8.65       9.59      12.71       8.71  ...   \n",
       "1255_g_at             5.50       5.50       5.50       5.50       5.50  ...   \n",
       "...                    ...        ...        ...        ...        ...  ...   \n",
       "AFFX-ThrX-5_at        5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-ThrX-M_at        5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-TrpnX-3_at       5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-TrpnX-5_at       5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-TrpnX-M_at       5.50       5.50       5.50       5.50       5.50  ...   \n",
       "\n",
       "                 GSM1094071  GSM1094072  GSM1094073  GSM1094074  GSM1094075  \\\n",
       "ID                                                                            \n",
       "1007_s_at            139.79       35.72       79.18      154.74      166.63   \n",
       "1053_at                8.32        8.79       11.64       17.01       17.19   \n",
       "117_at                24.91       11.09      309.80       42.61        9.57   \n",
       "121_at                 8.73       18.09        9.59        8.69        8.69   \n",
       "1255_g_at              5.50        5.50        5.50        5.50        5.50   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at         5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-ThrX-M_at         5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-TrpnX-3_at        5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-TrpnX-5_at        5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-TrpnX-M_at        5.50        5.50        5.50        5.50        5.50   \n",
       "\n",
       "                 GSM1094076  GSM1094077  GSM1094078  GSM1094079  GSM1094080  \n",
       "ID                                                                           \n",
       "1007_s_at             29.67       72.48       10.21      151.86      137.73  \n",
       "1053_at               25.84       16.13       43.64       11.46       22.56  \n",
       "117_at                38.21       50.33      321.67       14.87        9.55  \n",
       "121_at                12.21        8.64        8.64        8.68        8.69  \n",
       "1255_g_at              5.50        5.50        5.50        5.50        5.50  \n",
       "...                     ...         ...         ...         ...         ...  \n",
       "AFFX-ThrX-5_at         5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-ThrX-M_at         5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-TrpnX-3_at        5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-TrpnX-5_at        5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-TrpnX-M_at        5.50        5.50        5.50        5.50        5.50  \n",
       "\n",
       "[54675 rows x 48 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996730</th>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996731</th>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996732</th>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996733</th>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2996734</th>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2996735 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID  GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0              1007_s_at  U48705     NaN            Homo sapiens   \n",
       "1                1053_at  M87338     NaN            Homo sapiens   \n",
       "2                 117_at  X51757     NaN            Homo sapiens   \n",
       "3                 121_at  X69699     NaN            Homo sapiens   \n",
       "4              1255_g_at  L36861     NaN            Homo sapiens   \n",
       "...                  ...     ...     ...                     ...   \n",
       "2996730   AFFX-ThrX-5_at     5.5     NaN                     NaN   \n",
       "2996731   AFFX-ThrX-M_at     5.5     NaN                     NaN   \n",
       "2996732  AFFX-TrpnX-3_at     5.5     NaN                     NaN   \n",
       "2996733  AFFX-TrpnX-5_at     5.5     NaN                     NaN   \n",
       "2996734  AFFX-TrpnX-M_at     5.5     NaN                     NaN   \n",
       "\n",
       "        Annotation Date      Sequence Type                  Sequence Source  \\\n",
       "0           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "2           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "4           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                 ...                ...                              ...   \n",
       "2996730             NaN                NaN                              NaN   \n",
       "2996731             NaN                NaN                              NaN   \n",
       "2996732             NaN                NaN                              NaN   \n",
       "2996733             NaN                NaN                              NaN   \n",
       "2996734             NaN                NaN                              NaN   \n",
       "\n",
       "                                        Target Description  \\\n",
       "0        U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1        M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2        X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3        X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4        L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                    ...   \n",
       "2996730                                                NaN   \n",
       "2996731                                                NaN   \n",
       "2996732                                                NaN   \n",
       "2996733                                                NaN   \n",
       "2996734                                                NaN   \n",
       "\n",
       "        Representative Public ID  \\\n",
       "0                         U48705   \n",
       "1                         M87338   \n",
       "2                         X51757   \n",
       "3                         X69699   \n",
       "4                         L36861   \n",
       "...                          ...   \n",
       "2996730                      NaN   \n",
       "2996731                      NaN   \n",
       "2996732                      NaN   \n",
       "2996733                      NaN   \n",
       "2996734                      NaN   \n",
       "\n",
       "                                                Gene Title       Gene Symbol  \\\n",
       "0        discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1              replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2                     heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                             paired box 8              PAX8   \n",
       "4                  guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "...                                                    ...               ...   \n",
       "2996730                                                NaN               NaN   \n",
       "2996731                                                NaN               NaN   \n",
       "2996732                                                NaN               NaN   \n",
       "2996733                                                NaN               NaN   \n",
       "2996734                                                NaN               NaN   \n",
       "\n",
       "            ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0        780 /// 100616237  NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1                     5982  NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2                     3310                                          NM_002155   \n",
       "3                     7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                     2978                         NM_000409 /// XM_006715073   \n",
       "...                    ...                                                ...   \n",
       "2996730                NaN                                                NaN   \n",
       "2996731                NaN                                                NaN   \n",
       "2996732                NaN                                                NaN   \n",
       "2996733                NaN                                                NaN   \n",
       "2996734                NaN                                                NaN   \n",
       "\n",
       "                          Gene Ontology Biological Process  \\\n",
       "0        0001558 // regulation of cell growth // inferr...   \n",
       "1        0000278 // mitotic cell cycle // traceable aut...   \n",
       "2        0000902 // cell morphogenesis // inferred from...   \n",
       "3        0001655 // urogenital system development // in...   \n",
       "4        0007165 // signal transduction // non-traceabl...   \n",
       "...                                                    ...   \n",
       "2996730                                                NaN   \n",
       "2996731                                                NaN   \n",
       "2996732                                                NaN   \n",
       "2996733                                                NaN   \n",
       "2996734                                                NaN   \n",
       "\n",
       "                          Gene Ontology Cellular Component  \\\n",
       "0        0005576 // extracellular region // inferred fr...   \n",
       "1        0005634 // nucleus // inferred from electronic...   \n",
       "2        0005737 // cytoplasm // inferred from direct a...   \n",
       "3        0005634 // nucleus // inferred from direct ass...   \n",
       "4        0001750 // photoreceptor outer segment // infe...   \n",
       "...                                                    ...   \n",
       "2996730                                                NaN   \n",
       "2996731                                                NaN   \n",
       "2996732                                                NaN   \n",
       "2996733                                                NaN   \n",
       "2996734                                                NaN   \n",
       "\n",
       "                          Gene Ontology Molecular Function  \n",
       "0        0000166 // nucleotide binding // inferred from...  \n",
       "1        0000166 // nucleotide binding // inferred from...  \n",
       "2        0000166 // nucleotide binding // inferred from...  \n",
       "3        0000979 // RNA polymerase II core promoter seq...  \n",
       "4        0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                    ...  \n",
       "2996730                                                NaN  \n",
       "2996731                                                NaN  \n",
       "2996732                                                NaN  \n",
       "2996733                                                NaN  \n",
       "2996734                                                NaN  \n",
       "\n",
       "[2996735 rows x 16 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>1341.22</td>\n",
       "      <td>1980.88</td>\n",
       "      <td>1241.750</td>\n",
       "      <td>1284.08</td>\n",
       "      <td>455.520</td>\n",
       "      <td>15.020</td>\n",
       "      <td>8.490</td>\n",
       "      <td>2015.63</td>\n",
       "      <td>369.640</td>\n",
       "      <td>56.960</td>\n",
       "      <td>...</td>\n",
       "      <td>333.370</td>\n",
       "      <td>2458.34</td>\n",
       "      <td>180.17</td>\n",
       "      <td>736.710</td>\n",
       "      <td>460.15</td>\n",
       "      <td>878.59</td>\n",
       "      <td>460.98</td>\n",
       "      <td>8.25</td>\n",
       "      <td>161.38</td>\n",
       "      <td>526.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>7.83</td>\n",
       "      <td>10.19</td>\n",
       "      <td>13.070</td>\n",
       "      <td>11.64</td>\n",
       "      <td>76.920</td>\n",
       "      <td>11.640</td>\n",
       "      <td>23.720</td>\n",
       "      <td>13.40</td>\n",
       "      <td>5.810</td>\n",
       "      <td>23.780</td>\n",
       "      <td>...</td>\n",
       "      <td>5.770</td>\n",
       "      <td>9.76</td>\n",
       "      <td>7.01</td>\n",
       "      <td>5.500</td>\n",
       "      <td>8.54</td>\n",
       "      <td>6.73</td>\n",
       "      <td>7.71</td>\n",
       "      <td>10.09</td>\n",
       "      <td>11.64</td>\n",
       "      <td>11.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>7.83</td>\n",
       "      <td>10.19</td>\n",
       "      <td>13.070</td>\n",
       "      <td>11.64</td>\n",
       "      <td>76.920</td>\n",
       "      <td>11.640</td>\n",
       "      <td>23.720</td>\n",
       "      <td>13.40</td>\n",
       "      <td>5.810</td>\n",
       "      <td>23.780</td>\n",
       "      <td>...</td>\n",
       "      <td>5.770</td>\n",
       "      <td>9.76</td>\n",
       "      <td>7.01</td>\n",
       "      <td>5.500</td>\n",
       "      <td>8.54</td>\n",
       "      <td>6.73</td>\n",
       "      <td>7.71</td>\n",
       "      <td>10.09</td>\n",
       "      <td>11.64</td>\n",
       "      <td>11.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1P2</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.83</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC078883.4</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>6.83</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abParts</th>\n",
       "      <td>9.04</td>\n",
       "      <td>9.04</td>\n",
       "      <td>9.040</td>\n",
       "      <td>9.04</td>\n",
       "      <td>35.270</td>\n",
       "      <td>6313.010</td>\n",
       "      <td>233.550</td>\n",
       "      <td>9.04</td>\n",
       "      <td>9.040</td>\n",
       "      <td>9.040</td>\n",
       "      <td>...</td>\n",
       "      <td>9.040</td>\n",
       "      <td>9.04</td>\n",
       "      <td>9.05</td>\n",
       "      <td>9.040</td>\n",
       "      <td>11.09</td>\n",
       "      <td>9.61</td>\n",
       "      <td>9.04</td>\n",
       "      <td>52.44</td>\n",
       "      <td>43.59</td>\n",
       "      <td>9.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>13.290</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>44.860</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>av27s1</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.72</td>\n",
       "      <td>5.640</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>9.31</td>\n",
       "      <td>9.05</td>\n",
       "      <td>8.065</td>\n",
       "      <td>6.64</td>\n",
       "      <td>8.985</td>\n",
       "      <td>11.525</td>\n",
       "      <td>11.875</td>\n",
       "      <td>8.55</td>\n",
       "      <td>20.685</td>\n",
       "      <td>6.775</td>\n",
       "      <td>...</td>\n",
       "      <td>5.625</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mir-223</th>\n",
       "      <td>26.66</td>\n",
       "      <td>7.34</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>6.250</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25851 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GSM493251  GSM493252  GSM493253  GSM493254  GSM493255  \\\n",
       "Gene                                                                   \n",
       " ABCB4           1341.22    1980.88   1241.750    1284.08    455.520   \n",
       " ABCC6P1            7.83      10.19     13.070      11.64     76.920   \n",
       " ABCC6P2            7.83      10.19     13.070      11.64     76.920   \n",
       " ABCD1P2            5.50       5.83      5.500       5.50      5.500   \n",
       " AC078883.4         5.50       5.50      5.500       5.50      5.500   \n",
       "...                  ...        ...        ...        ...        ...   \n",
       "abParts             9.04       9.04      9.040       9.04     35.270   \n",
       "alpha               5.50       5.50      5.500       5.50      5.500   \n",
       "av27s1              5.50       5.72      5.640       5.50      5.500   \n",
       "hsa-let-7a-3        9.31       9.05      8.065       6.64      8.985   \n",
       "mir-223            26.66       7.34      5.500       5.50      5.500   \n",
       "\n",
       "               GSM493256  GSM493257  GSM493258  GSM493259  GSM493260  ...  \\\n",
       "Gene                                                                  ...   \n",
       " ABCB4            15.020      8.490    2015.63    369.640     56.960  ...   \n",
       " ABCC6P1          11.640     23.720      13.40      5.810     23.780  ...   \n",
       " ABCC6P2          11.640     23.720      13.40      5.810     23.780  ...   \n",
       " ABCD1P2           5.500      5.500       5.50      5.500      5.500  ...   \n",
       " AC078883.4        5.500      5.500       5.50      5.500      5.500  ...   \n",
       "...                  ...        ...        ...        ...        ...  ...   \n",
       "abParts         6313.010    233.550       9.04      9.040      9.040  ...   \n",
       "alpha             13.290      5.500       5.50     44.860      5.500  ...   \n",
       "av27s1             5.500      5.500       5.50      5.500      5.500  ...   \n",
       "hsa-let-7a-3      11.525     11.875       8.55     20.685      6.775  ...   \n",
       "mir-223            6.250      5.500       5.50      5.500      5.500  ...   \n",
       "\n",
       "               GSM1094071  GSM1094072  GSM1094073  GSM1094074  GSM1094075  \\\n",
       "Gene                                                                        \n",
       " ABCB4            333.370     2458.34      180.17     736.710      460.15   \n",
       " ABCC6P1            5.770        9.76        7.01       5.500        8.54   \n",
       " ABCC6P2            5.770        9.76        7.01       5.500        8.54   \n",
       " ABCD1P2            5.500        5.50        5.50       5.500        5.50   \n",
       " AC078883.4         5.500        5.50        5.50       5.500        5.50   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts             9.040        9.04        9.05       9.040       11.09   \n",
       "alpha               5.500        5.50        5.50       5.500        5.50   \n",
       "av27s1              5.500        5.50        5.50       5.500        5.50   \n",
       "hsa-let-7a-3        5.625        5.50        5.50       5.675        5.50   \n",
       "mir-223             5.500        5.50        5.50       5.500        5.50   \n",
       "\n",
       "               GSM1094076  GSM1094077  GSM1094078  GSM1094079  GSM1094080  \n",
       "Gene                                                                       \n",
       " ABCB4             878.59      460.98        8.25      161.38      526.21  \n",
       " ABCC6P1             6.73        7.71       10.09       11.64       11.64  \n",
       " ABCC6P2             6.73        7.71       10.09       11.64       11.64  \n",
       " ABCD1P2             5.50        5.50        5.50        5.50        5.50  \n",
       " AC078883.4          6.83        5.50        5.50        5.50        5.50  \n",
       "...                   ...         ...         ...         ...         ...  \n",
       "abParts              9.61        9.04       52.44       43.59        9.04  \n",
       "alpha                5.50        5.50        5.50        5.50        5.50  \n",
       "av27s1               5.50        5.50        5.50        5.50        5.50  \n",
       "hsa-let-7a-3         5.50        5.50        5.50        5.50        5.50  \n",
       "mir-223              5.50        5.50        5.50        5.50        5.50  \n",
       "\n",
       "[25851 rows x 48 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.640</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>18.570000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>6.410000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.980000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>8.410000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>7.950000</td>\n",
       "      <td>12.980</td>\n",
       "      <td>13.340000</td>\n",
       "      <td>8.350000</td>\n",
       "      <td>8.840000</td>\n",
       "      <td>7.170000</td>\n",
       "      <td>10.140000</td>\n",
       "      <td>6.800000</td>\n",
       "      <td>6.540000</td>\n",
       "      <td>15.120000</td>\n",
       "      <td>...</td>\n",
       "      <td>10.540000</td>\n",
       "      <td>8.270000</td>\n",
       "      <td>16.220000</td>\n",
       "      <td>7.580000</td>\n",
       "      <td>9.610000</td>\n",
       "      <td>13.200000</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>39.740000</td>\n",
       "      <td>7.510000</td>\n",
       "      <td>6.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>7.120000</td>\n",
       "      <td>5.500</td>\n",
       "      <td>6.805000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.930000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>6.555000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>3303.540000</td>\n",
       "      <td>1751.345</td>\n",
       "      <td>4304.375000</td>\n",
       "      <td>4467.760000</td>\n",
       "      <td>1225.825000</td>\n",
       "      <td>404.895000</td>\n",
       "      <td>276.810000</td>\n",
       "      <td>922.500000</td>\n",
       "      <td>184.890000</td>\n",
       "      <td>296.650000</td>\n",
       "      <td>...</td>\n",
       "      <td>503.735000</td>\n",
       "      <td>832.570000</td>\n",
       "      <td>1589.435000</td>\n",
       "      <td>924.025000</td>\n",
       "      <td>467.175000</td>\n",
       "      <td>984.970000</td>\n",
       "      <td>678.560000</td>\n",
       "      <td>249.025000</td>\n",
       "      <td>1012.500000</td>\n",
       "      <td>963.050000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>13.250000</td>\n",
       "      <td>10.690</td>\n",
       "      <td>33.790000</td>\n",
       "      <td>23.860000</td>\n",
       "      <td>27.390000</td>\n",
       "      <td>9.860000</td>\n",
       "      <td>8.260000</td>\n",
       "      <td>49.400000</td>\n",
       "      <td>9.820000</td>\n",
       "      <td>14.360000</td>\n",
       "      <td>...</td>\n",
       "      <td>17.580000</td>\n",
       "      <td>24.270000</td>\n",
       "      <td>27.740000</td>\n",
       "      <td>16.180000</td>\n",
       "      <td>8.800000</td>\n",
       "      <td>16.650000</td>\n",
       "      <td>19.140000</td>\n",
       "      <td>9.390000</td>\n",
       "      <td>21.520000</td>\n",
       "      <td>18.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.680000</td>\n",
       "      <td>41.110000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>121.390000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>6.110000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>120.173333</td>\n",
       "      <td>135.650</td>\n",
       "      <td>144.273333</td>\n",
       "      <td>160.646667</td>\n",
       "      <td>192.403333</td>\n",
       "      <td>69.213333</td>\n",
       "      <td>164.600000</td>\n",
       "      <td>153.056667</td>\n",
       "      <td>118.900000</td>\n",
       "      <td>124.290000</td>\n",
       "      <td>...</td>\n",
       "      <td>52.633333</td>\n",
       "      <td>102.250000</td>\n",
       "      <td>73.350000</td>\n",
       "      <td>83.226667</td>\n",
       "      <td>84.833333</td>\n",
       "      <td>115.490000</td>\n",
       "      <td>64.146667</td>\n",
       "      <td>122.486667</td>\n",
       "      <td>66.980000</td>\n",
       "      <td>111.153333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>109.345000</td>\n",
       "      <td>103.390</td>\n",
       "      <td>115.915000</td>\n",
       "      <td>113.375000</td>\n",
       "      <td>33.390000</td>\n",
       "      <td>169.690000</td>\n",
       "      <td>97.565000</td>\n",
       "      <td>68.930000</td>\n",
       "      <td>105.195000</td>\n",
       "      <td>18.495000</td>\n",
       "      <td>...</td>\n",
       "      <td>18.360000</td>\n",
       "      <td>28.040000</td>\n",
       "      <td>86.410000</td>\n",
       "      <td>41.670000</td>\n",
       "      <td>31.345000</td>\n",
       "      <td>24.975000</td>\n",
       "      <td>43.145000</td>\n",
       "      <td>16.535000</td>\n",
       "      <td>16.925000</td>\n",
       "      <td>19.220000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>46.093333</td>\n",
       "      <td>37.000</td>\n",
       "      <td>36.803333</td>\n",
       "      <td>34.036667</td>\n",
       "      <td>53.216667</td>\n",
       "      <td>21.713333</td>\n",
       "      <td>37.506667</td>\n",
       "      <td>89.816667</td>\n",
       "      <td>77.906667</td>\n",
       "      <td>36.686667</td>\n",
       "      <td>...</td>\n",
       "      <td>58.003333</td>\n",
       "      <td>15.563333</td>\n",
       "      <td>8.223333</td>\n",
       "      <td>11.126667</td>\n",
       "      <td>9.826667</td>\n",
       "      <td>10.496667</td>\n",
       "      <td>13.913333</td>\n",
       "      <td>11.370000</td>\n",
       "      <td>11.766667</td>\n",
       "      <td>15.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>54.605000</td>\n",
       "      <td>61.375</td>\n",
       "      <td>60.770000</td>\n",
       "      <td>61.425000</td>\n",
       "      <td>89.515000</td>\n",
       "      <td>119.045000</td>\n",
       "      <td>242.765000</td>\n",
       "      <td>115.675000</td>\n",
       "      <td>71.390000</td>\n",
       "      <td>100.650000</td>\n",
       "      <td>...</td>\n",
       "      <td>80.365000</td>\n",
       "      <td>107.690000</td>\n",
       "      <td>74.930000</td>\n",
       "      <td>66.605000</td>\n",
       "      <td>115.315000</td>\n",
       "      <td>64.875000</td>\n",
       "      <td>54.920000</td>\n",
       "      <td>206.205000</td>\n",
       "      <td>62.815000</td>\n",
       "      <td>64.870000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM493251  GSM493252    GSM493253    GSM493254    GSM493255  \\\n",
       "A1BG         5.500000      5.640     5.500000     5.500000     5.500000   \n",
       "A1BG-AS1     7.950000     12.980    13.340000     8.350000     8.840000   \n",
       "A1CF         7.120000      5.500     6.805000     5.500000     5.500000   \n",
       "A2M       3303.540000   1751.345  4304.375000  4467.760000  1225.825000   \n",
       "A2M-AS1     13.250000     10.690    33.790000    23.860000    27.390000   \n",
       "...               ...        ...          ...          ...          ...   \n",
       "ZYG11A       5.500000      5.500     5.500000     5.500000     5.500000   \n",
       "ZYG11B     120.173333    135.650   144.273333   160.646667   192.403333   \n",
       "ZYX        109.345000    103.390   115.915000   113.375000    33.390000   \n",
       "ZZEF1       46.093333     37.000    36.803333    34.036667    53.216667   \n",
       "ZZZ3        54.605000     61.375    60.770000    61.425000    89.515000   \n",
       "\n",
       "           GSM493256   GSM493257   GSM493258   GSM493259   GSM493260  ...  \\\n",
       "A1BG        5.500000   18.570000    5.500000    5.500000    5.500000  ...   \n",
       "A1BG-AS1    7.170000   10.140000    6.800000    6.540000   15.120000  ...   \n",
       "A1CF        5.500000    5.930000    5.500000    6.555000    5.500000  ...   \n",
       "A2M       404.895000  276.810000  922.500000  184.890000  296.650000  ...   \n",
       "A2M-AS1     9.860000    8.260000   49.400000    9.820000   14.360000  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A      5.680000   41.110000    5.500000    5.500000    5.500000  ...   \n",
       "ZYG11B     69.213333  164.600000  153.056667  118.900000  124.290000  ...   \n",
       "ZYX       169.690000   97.565000   68.930000  105.195000   18.495000  ...   \n",
       "ZZEF1      21.713333   37.506667   89.816667   77.906667   36.686667  ...   \n",
       "ZZZ3      119.045000  242.765000  115.675000   71.390000  100.650000  ...   \n",
       "\n",
       "          GSM1094071  GSM1094072   GSM1094073  GSM1094074  GSM1094075  \\\n",
       "A1BG        5.500000    5.500000     6.410000    5.500000    5.500000   \n",
       "A1BG-AS1   10.540000    8.270000    16.220000    7.580000    9.610000   \n",
       "A1CF        5.500000    5.500000     5.500000    5.500000    5.500000   \n",
       "A2M       503.735000  832.570000  1589.435000  924.025000  467.175000   \n",
       "A2M-AS1    17.580000   24.270000    27.740000   16.180000    8.800000   \n",
       "...              ...         ...          ...         ...         ...   \n",
       "ZYG11A      5.500000    5.500000     5.500000    5.500000    5.500000   \n",
       "ZYG11B     52.633333  102.250000    73.350000   83.226667   84.833333   \n",
       "ZYX        18.360000   28.040000    86.410000   41.670000   31.345000   \n",
       "ZZEF1      58.003333   15.563333     8.223333   11.126667    9.826667   \n",
       "ZZZ3       80.365000  107.690000    74.930000   66.605000  115.315000   \n",
       "\n",
       "          GSM1094076  GSM1094077  GSM1094078   GSM1094079  GSM1094080  \n",
       "A1BG        5.980000    5.500000    8.410000     5.500000    5.500000  \n",
       "A1BG-AS1   13.200000    6.500000   39.740000     7.510000    6.300000  \n",
       "A1CF        5.500000    5.500000    5.500000     5.500000    5.500000  \n",
       "A2M       984.970000  678.560000  249.025000  1012.500000  963.050000  \n",
       "A2M-AS1    16.650000   19.140000    9.390000    21.520000   18.600000  \n",
       "...              ...         ...         ...          ...         ...  \n",
       "ZYG11A      5.500000    5.500000  121.390000     5.500000    6.110000  \n",
       "ZYG11B    115.490000   64.146667  122.486667    66.980000  111.153333  \n",
       "ZYX        24.975000   43.145000   16.535000    16.925000   19.220000  \n",
       "ZZEF1      10.496667   13.913333   11.370000    11.766667   15.500000  \n",
       "ZZZ3       64.875000   54.920000  206.205000    62.815000   64.870000  \n",
       "\n",
       "[17857 rows x 48 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:718: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n"
     ]
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM493265</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.53</td>\n",
       "      <td>27.98</td>\n",
       "      <td>5.5</td>\n",
       "      <td>630.915</td>\n",
       "      <td>14.83</td>\n",
       "      <td>5.84</td>\n",
       "      <td>5.79</td>\n",
       "      <td>...</td>\n",
       "      <td>21.53</td>\n",
       "      <td>86.84</td>\n",
       "      <td>8.83</td>\n",
       "      <td>6.735</td>\n",
       "      <td>36.676667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>98.686667</td>\n",
       "      <td>34.93</td>\n",
       "      <td>30.373333</td>\n",
       "      <td>71.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM493270</th>\n",
       "      <td>1.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.77</td>\n",
       "      <td>5.5</td>\n",
       "      <td>968.32</td>\n",
       "      <td>34.95</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5.79</td>\n",
       "      <td>...</td>\n",
       "      <td>30.695</td>\n",
       "      <td>90.66</td>\n",
       "      <td>13.94</td>\n",
       "      <td>11.26</td>\n",
       "      <td>33.223333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>158.316667</td>\n",
       "      <td>70.065</td>\n",
       "      <td>61.263333</td>\n",
       "      <td>68.105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094056</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>9.68</td>\n",
       "      <td>5.5</td>\n",
       "      <td>380.405</td>\n",
       "      <td>9.32</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5.84</td>\n",
       "      <td>...</td>\n",
       "      <td>48.445</td>\n",
       "      <td>191.94</td>\n",
       "      <td>12.93</td>\n",
       "      <td>9.78</td>\n",
       "      <td>41.331667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>89.596667</td>\n",
       "      <td>15.265</td>\n",
       "      <td>16.493333</td>\n",
       "      <td>108.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094057</th>\n",
       "      <td>1.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.95</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1465.11</td>\n",
       "      <td>27.9</td>\n",
       "      <td>5.5</td>\n",
       "      <td>154.36</td>\n",
       "      <td>...</td>\n",
       "      <td>48.365</td>\n",
       "      <td>281.2</td>\n",
       "      <td>7.29</td>\n",
       "      <td>7.32</td>\n",
       "      <td>46.606667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>50.603333</td>\n",
       "      <td>28.48</td>\n",
       "      <td>10.89</td>\n",
       "      <td>58.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094060</th>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.95</td>\n",
       "      <td>5.5</td>\n",
       "      <td>680.91</td>\n",
       "      <td>14.43</td>\n",
       "      <td>5.5</td>\n",
       "      <td>12.75</td>\n",
       "      <td>...</td>\n",
       "      <td>47.235</td>\n",
       "      <td>28.79</td>\n",
       "      <td>10.14</td>\n",
       "      <td>11.98</td>\n",
       "      <td>45.533333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>70.15</td>\n",
       "      <td>15.255</td>\n",
       "      <td>23.983333</td>\n",
       "      <td>101.235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094061</th>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.29</td>\n",
       "      <td>5.5</td>\n",
       "      <td>999.32</td>\n",
       "      <td>11.21</td>\n",
       "      <td>5.5</td>\n",
       "      <td>18.01</td>\n",
       "      <td>...</td>\n",
       "      <td>32.09</td>\n",
       "      <td>81.36</td>\n",
       "      <td>19.22</td>\n",
       "      <td>9.825</td>\n",
       "      <td>28.378333</td>\n",
       "      <td>5.56</td>\n",
       "      <td>76.246667</td>\n",
       "      <td>54.955</td>\n",
       "      <td>10.566667</td>\n",
       "      <td>67.585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094063</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.51</td>\n",
       "      <td>5.5</td>\n",
       "      <td>474.915</td>\n",
       "      <td>12.02</td>\n",
       "      <td>5.5</td>\n",
       "      <td>46.27</td>\n",
       "      <td>...</td>\n",
       "      <td>41.44</td>\n",
       "      <td>61.6</td>\n",
       "      <td>23.76</td>\n",
       "      <td>18.295</td>\n",
       "      <td>43.958333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>87.03</td>\n",
       "      <td>23.975</td>\n",
       "      <td>10.793333</td>\n",
       "      <td>94.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094066</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.71</td>\n",
       "      <td>22.0</td>\n",
       "      <td>5.945</td>\n",
       "      <td>365.945</td>\n",
       "      <td>11.49</td>\n",
       "      <td>5.56</td>\n",
       "      <td>60.55</td>\n",
       "      <td>...</td>\n",
       "      <td>35.47</td>\n",
       "      <td>65.86</td>\n",
       "      <td>32.64</td>\n",
       "      <td>25.8</td>\n",
       "      <td>55.596667</td>\n",
       "      <td>7.74</td>\n",
       "      <td>89.68</td>\n",
       "      <td>12.18</td>\n",
       "      <td>14.486667</td>\n",
       "      <td>95.235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094067</th>\n",
       "      <td>0.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.2</td>\n",
       "      <td>6.55</td>\n",
       "      <td>1191.28</td>\n",
       "      <td>23.69</td>\n",
       "      <td>5.5</td>\n",
       "      <td>9.22</td>\n",
       "      <td>...</td>\n",
       "      <td>33.985</td>\n",
       "      <td>37.59</td>\n",
       "      <td>24.58</td>\n",
       "      <td>22.49</td>\n",
       "      <td>27.738333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>109.716667</td>\n",
       "      <td>42.44</td>\n",
       "      <td>10.15</td>\n",
       "      <td>61.765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094071</th>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>10.54</td>\n",
       "      <td>5.5</td>\n",
       "      <td>503.735</td>\n",
       "      <td>17.58</td>\n",
       "      <td>5.5</td>\n",
       "      <td>93.73</td>\n",
       "      <td>...</td>\n",
       "      <td>53.105</td>\n",
       "      <td>159.57</td>\n",
       "      <td>14.95</td>\n",
       "      <td>11.98</td>\n",
       "      <td>41.068333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>52.633333</td>\n",
       "      <td>18.36</td>\n",
       "      <td>58.003333</td>\n",
       "      <td>80.365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094074</th>\n",
       "      <td>1.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.58</td>\n",
       "      <td>5.5</td>\n",
       "      <td>924.025</td>\n",
       "      <td>16.18</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.8</td>\n",
       "      <td>...</td>\n",
       "      <td>64.32</td>\n",
       "      <td>139.03</td>\n",
       "      <td>20.97</td>\n",
       "      <td>19.525</td>\n",
       "      <td>33.635</td>\n",
       "      <td>5.5</td>\n",
       "      <td>83.226667</td>\n",
       "      <td>41.67</td>\n",
       "      <td>11.126667</td>\n",
       "      <td>66.605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094075</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>9.61</td>\n",
       "      <td>5.5</td>\n",
       "      <td>467.175</td>\n",
       "      <td>8.8</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.38</td>\n",
       "      <td>...</td>\n",
       "      <td>38.765</td>\n",
       "      <td>92.65</td>\n",
       "      <td>49.51</td>\n",
       "      <td>48.36</td>\n",
       "      <td>31.856667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>84.833333</td>\n",
       "      <td>31.345</td>\n",
       "      <td>9.826667</td>\n",
       "      <td>115.315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094078</th>\n",
       "      <td>1.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.41</td>\n",
       "      <td>39.74</td>\n",
       "      <td>5.5</td>\n",
       "      <td>249.025</td>\n",
       "      <td>9.39</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5.97</td>\n",
       "      <td>...</td>\n",
       "      <td>179.07</td>\n",
       "      <td>240.64</td>\n",
       "      <td>17.37</td>\n",
       "      <td>22.465</td>\n",
       "      <td>46.6</td>\n",
       "      <td>121.39</td>\n",
       "      <td>122.486667</td>\n",
       "      <td>16.535</td>\n",
       "      <td>11.37</td>\n",
       "      <td>206.205</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows Ã— 17860 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer   Age Gender   A1BG A1BG-AS1   A1CF      A2M  \\\n",
       "GSM493265                    1.0  37.0    1.0  29.53    27.98    5.5  630.915   \n",
       "GSM493270                    1.0  58.0    1.0    5.5     6.77    5.5   968.32   \n",
       "GSM1094056                   0.0  20.0    1.0    5.5     9.68    5.5  380.405   \n",
       "GSM1094057                   1.0  68.0    1.0    5.5     7.95    5.5  1465.11   \n",
       "GSM1094060                   0.0  32.0    1.0    5.5     7.95    5.5   680.91   \n",
       "GSM1094061                   0.0  43.0    0.0    5.5     6.29    5.5   999.32   \n",
       "GSM1094063                   1.0  40.0    0.0    5.5     7.51    5.5  474.915   \n",
       "GSM1094066                   0.0  27.0    1.0  10.71     22.0  5.945  365.945   \n",
       "GSM1094067                   0.0  70.0    0.0    5.5      7.2   6.55  1191.28   \n",
       "GSM1094071                   0.0  57.0    0.0    5.5    10.54    5.5  503.735   \n",
       "GSM1094074                   1.0  55.0    1.0    5.5     7.58    5.5  924.025   \n",
       "GSM1094075                   1.0  51.0    1.0    5.5     9.61    5.5  467.175   \n",
       "GSM1094078                   1.0  63.0    0.0   8.41    39.74    5.5  249.025   \n",
       "\n",
       "           A2M-AS1 A2ML1   A2MP1  ...  ZWILCH   ZWINT   ZXDA    ZXDB  \\\n",
       "GSM493265    14.83  5.84    5.79  ...   21.53   86.84   8.83   6.735   \n",
       "GSM493270    34.95   5.5    5.79  ...  30.695   90.66  13.94   11.26   \n",
       "GSM1094056    9.32   5.5    5.84  ...  48.445  191.94  12.93    9.78   \n",
       "GSM1094057    27.9   5.5  154.36  ...  48.365   281.2   7.29    7.32   \n",
       "GSM1094060   14.43   5.5   12.75  ...  47.235   28.79  10.14   11.98   \n",
       "GSM1094061   11.21   5.5   18.01  ...   32.09   81.36  19.22   9.825   \n",
       "GSM1094063   12.02   5.5   46.27  ...   41.44    61.6  23.76  18.295   \n",
       "GSM1094066   11.49  5.56   60.55  ...   35.47   65.86  32.64    25.8   \n",
       "GSM1094067   23.69   5.5    9.22  ...  33.985   37.59  24.58   22.49   \n",
       "GSM1094071   17.58   5.5   93.73  ...  53.105  159.57  14.95   11.98   \n",
       "GSM1094074   16.18   5.5     6.8  ...   64.32  139.03  20.97  19.525   \n",
       "GSM1094075     8.8   5.5    6.38  ...  38.765   92.65  49.51   48.36   \n",
       "GSM1094078    9.39   5.5    5.97  ...  179.07  240.64  17.37  22.465   \n",
       "\n",
       "                 ZXDC  ZYG11A      ZYG11B     ZYX      ZZEF1     ZZZ3  \n",
       "GSM493265   36.676667     5.5   98.686667   34.93  30.373333    71.49  \n",
       "GSM493270   33.223333     5.5  158.316667  70.065  61.263333   68.105  \n",
       "GSM1094056  41.331667     5.5   89.596667  15.265  16.493333   108.55  \n",
       "GSM1094057  46.606667     5.5   50.603333   28.48      10.89    58.92  \n",
       "GSM1094060  45.533333     5.5       70.15  15.255  23.983333  101.235  \n",
       "GSM1094061  28.378333    5.56   76.246667  54.955  10.566667   67.585  \n",
       "GSM1094063  43.958333     5.5       87.03  23.975  10.793333    94.43  \n",
       "GSM1094066  55.596667    7.74       89.68   12.18  14.486667   95.235  \n",
       "GSM1094067  27.738333     5.5  109.716667   42.44      10.15   61.765  \n",
       "GSM1094071  41.068333     5.5   52.633333   18.36  58.003333   80.365  \n",
       "GSM1094074     33.635     5.5   83.226667   41.67  11.126667   66.605  \n",
       "GSM1094075  31.856667     5.5   84.833333  31.345   9.826667  115.315  \n",
       "GSM1094078       46.6  121.39  122.486667  16.535      11.37  206.205  \n",
       "\n",
       "[13 rows x 17860 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 13 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '0.0' with 6 occurrences. This represents 46.15% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 37.0\n",
      "  50% (Median): 51.0\n",
      "  75%: 58.0\n",
      "Min: 20.0\n",
      "Max: 70.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 5 occurrences. This represents 38.46% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"SNP array profiling of childhood adrenocortical tumours reveals distinct pathways of tumourigenesis and highlights candidate driver genes\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM610697</th>\n",
       "      <th>GSM610698</th>\n",
       "      <th>GSM610699</th>\n",
       "      <th>GSM610700</th>\n",
       "      <th>GSM610701</th>\n",
       "      <th>GSM610702</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Male</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age  (months): 42.96</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "5  !Sample_characteristics_ch1   \n",
       "6  !Sample_characteristics_ch1   \n",
       "7  !Sample_characteristics_ch1   \n",
       "8  !Sample_characteristics_ch1   \n",
       "9  !Sample_characteristics_ch1   \n",
       "\n",
       "                                           GSM610697  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2                               age  (months): 42.96   \n",
       "3  histological type: Paired blood sample corresp...   \n",
       "4                                            tnm: --   \n",
       "5                                  tp53 mutation: --   \n",
       "6                                   virilization: --   \n",
       "7                              cushing  syndrome: --   \n",
       "8                                      treatment: --   \n",
       "9                             follow-up (months): --   \n",
       "\n",
       "                                           GSM610698  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610699  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610700  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                       genger: Male   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610701  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610702  \n",
       "0                        geographical origin: Brazil  \n",
       "1                                     genger: Female  \n",
       "2  histological type: Paired blood sample corresp...  \n",
       "3                                            tnm: --  \n",
       "4                                  tp53 mutation: --  \n",
       "5                                   virilization: --  \n",
       "6                              cushing  syndrome: --  \n",
       "7                                      treatment: --  \n",
       "8                             follow-up (months): --  \n",
       "9                                                NaN  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stop: No trait\n",
    "\n",
    "cohort = accession_num = \"GSE35066\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Study of miRNA expression in childhood adrenocortical tumors\"\n",
      "!Series_summary\t\"We studied the miRNA expression profile of a series of childhood adrenocortical tumors (ACT) and age-matched normal adrenal samples\"\n",
      "!Series_overall_design\t\"25 ACT - 5 normal\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM495879</th>\n",
       "      <th>GSM495880</th>\n",
       "      <th>GSM495881</th>\n",
       "      <th>GSM495882</th>\n",
       "      <th>GSM495883</th>\n",
       "      <th>GSM495884</th>\n",
       "      <th>GSM495885</th>\n",
       "      <th>GSM495886</th>\n",
       "      <th>GSM495887</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM495899</th>\n",
       "      <th>GSM495900</th>\n",
       "      <th>GSM495901</th>\n",
       "      <th>GSM495902</th>\n",
       "      <th>GSM495903</th>\n",
       "      <th>GSM495904</th>\n",
       "      <th>GSM495905</th>\n",
       "      <th>GSM495906</th>\n",
       "      <th>GSM495907</th>\n",
       "      <th>GSM495908</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "      <td>tissue: Adrenal cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: adrenocortical tumor</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                            GSM495879  \\\n",
       "0  !Sample_characteristics_ch1               tissue: Adrenal cortex   \n",
       "1  !Sample_characteristics_ch1  disease state: adrenocortical tumor   \n",
       "\n",
       "                             GSM495880                            GSM495881  \\\n",
       "0               tissue: Adrenal cortex               tissue: Adrenal cortex   \n",
       "1  disease state: adrenocortical tumor  disease state: adrenocortical tumor   \n",
       "\n",
       "                             GSM495882                            GSM495883  \\\n",
       "0               tissue: Adrenal cortex               tissue: Adrenal cortex   \n",
       "1  disease state: adrenocortical tumor  disease state: adrenocortical tumor   \n",
       "\n",
       "                             GSM495884                            GSM495885  \\\n",
       "0               tissue: Adrenal cortex               tissue: Adrenal cortex   \n",
       "1  disease state: adrenocortical tumor  disease state: adrenocortical tumor   \n",
       "\n",
       "                             GSM495886                            GSM495887  \\\n",
       "0               tissue: Adrenal cortex               tissue: Adrenal cortex   \n",
       "1  disease state: adrenocortical tumor  disease state: adrenocortical tumor   \n",
       "\n",
       "   ...                            GSM495899  \\\n",
       "0  ...               tissue: Adrenal cortex   \n",
       "1  ...  disease state: adrenocortical tumor   \n",
       "\n",
       "                             GSM495900                            GSM495901  \\\n",
       "0               tissue: Adrenal cortex               tissue: Adrenal cortex   \n",
       "1  disease state: adrenocortical tumor  disease state: adrenocortical tumor   \n",
       "\n",
       "                             GSM495902                            GSM495903  \\\n",
       "0               tissue: Adrenal cortex               tissue: Adrenal cortex   \n",
       "1  disease state: adrenocortical tumor  disease state: adrenocortical tumor   \n",
       "\n",
       "                GSM495904               GSM495905               GSM495906  \\\n",
       "0  tissue: Adrenal cortex  tissue: Adrenal cortex  tissue: Adrenal cortex   \n",
       "1   disease state: normal   disease state: normal   disease state: normal   \n",
       "\n",
       "                GSM495907               GSM495908  \n",
       "0  tissue: Adrenal cortex  tissue: Adrenal cortex  \n",
       "1   disease state: normal   disease state: normal  \n",
       "\n",
       "[2 rows x 31 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stop: No gene symbol (Pure miRNA data is not suitable)\n",
    "cohort = accession_num = \"GSE19856\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'disease state: adrenocortical tumor', 'disease state: normal'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease state: adrenocortical tumor':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM495879</th>\n",
       "      <th>GSM495880</th>\n",
       "      <th>GSM495881</th>\n",
       "      <th>GSM495882</th>\n",
       "      <th>GSM495883</th>\n",
       "      <th>GSM495884</th>\n",
       "      <th>GSM495885</th>\n",
       "      <th>GSM495886</th>\n",
       "      <th>GSM495887</th>\n",
       "      <th>GSM495888</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM495899</th>\n",
       "      <th>GSM495900</th>\n",
       "      <th>GSM495901</th>\n",
       "      <th>GSM495902</th>\n",
       "      <th>GSM495903</th>\n",
       "      <th>GSM495904</th>\n",
       "      <th>GSM495905</th>\n",
       "      <th>GSM495906</th>\n",
       "      <th>GSM495907</th>\n",
       "      <th>GSM495908</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM495879  GSM495880  GSM495881  GSM495882  GSM495883  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "\n",
       "                       GSM495884  GSM495885  GSM495886  GSM495887  GSM495888  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "\n",
       "                       ...  GSM495899  GSM495900  GSM495901  GSM495902  \\\n",
       "Adrenocortical Cancer  ...          1          1          1          1   \n",
       "\n",
       "                       GSM495903  GSM495904  GSM495905  GSM495906  GSM495907  \\\n",
       "Adrenocortical Cancer          1          0          0          0          0   \n",
       "\n",
       "                       GSM495908  \n",
       "Adrenocortical Cancer          0  \n",
       "\n",
       "[1 rows x 30 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM495879</th>\n",
       "      <th>GSM495880</th>\n",
       "      <th>GSM495881</th>\n",
       "      <th>GSM495882</th>\n",
       "      <th>GSM495883</th>\n",
       "      <th>GSM495884</th>\n",
       "      <th>GSM495885</th>\n",
       "      <th>GSM495886</th>\n",
       "      <th>GSM495887</th>\n",
       "      <th>GSM495888</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM495899</th>\n",
       "      <th>GSM495900</th>\n",
       "      <th>GSM495901</th>\n",
       "      <th>GSM495902</th>\n",
       "      <th>GSM495903</th>\n",
       "      <th>GSM495904</th>\n",
       "      <th>GSM495905</th>\n",
       "      <th>GSM495906</th>\n",
       "      <th>GSM495907</th>\n",
       "      <th>GSM495908</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DarkCorner</th>\n",
       "      <td>-7.21640</td>\n",
       "      <td>-6.898350</td>\n",
       "      <td>-9.59862</td>\n",
       "      <td>-4.718430</td>\n",
       "      <td>-9.114260</td>\n",
       "      <td>-5.497120</td>\n",
       "      <td>-8.27575</td>\n",
       "      <td>-0.418281</td>\n",
       "      <td>-7.752440</td>\n",
       "      <td>5.263860</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.673030</td>\n",
       "      <td>-10.034100</td>\n",
       "      <td>-2.87338</td>\n",
       "      <td>-6.890780</td>\n",
       "      <td>-13.643800</td>\n",
       "      <td>-11.24230</td>\n",
       "      <td>-14.145600</td>\n",
       "      <td>-13.803400</td>\n",
       "      <td>-11.586800</td>\n",
       "      <td>-11.498000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dmr_285</th>\n",
       "      <td>-3.24370</td>\n",
       "      <td>0.029345</td>\n",
       "      <td>-2.96027</td>\n",
       "      <td>-3.947260</td>\n",
       "      <td>-2.508740</td>\n",
       "      <td>-2.449810</td>\n",
       "      <td>-2.76578</td>\n",
       "      <td>-2.222550</td>\n",
       "      <td>-2.485020</td>\n",
       "      <td>0.194956</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.622680</td>\n",
       "      <td>-3.282510</td>\n",
       "      <td>-1.49847</td>\n",
       "      <td>-3.208890</td>\n",
       "      <td>-1.656840</td>\n",
       "      <td>-2.42992</td>\n",
       "      <td>-2.638710</td>\n",
       "      <td>-0.923110</td>\n",
       "      <td>-1.950690</td>\n",
       "      <td>1.098530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dmr_3</th>\n",
       "      <td>-4.14763</td>\n",
       "      <td>-2.179640</td>\n",
       "      <td>-1.86509</td>\n",
       "      <td>-2.264490</td>\n",
       "      <td>-1.913330</td>\n",
       "      <td>-3.938300</td>\n",
       "      <td>-2.03988</td>\n",
       "      <td>-0.926865</td>\n",
       "      <td>1.122930</td>\n",
       "      <td>-0.061020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.368513</td>\n",
       "      <td>1.808110</td>\n",
       "      <td>2.45639</td>\n",
       "      <td>0.092283</td>\n",
       "      <td>-0.197860</td>\n",
       "      <td>1.94856</td>\n",
       "      <td>0.776828</td>\n",
       "      <td>4.097750</td>\n",
       "      <td>0.194788</td>\n",
       "      <td>2.004070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dmr_308</th>\n",
       "      <td>-2.56221</td>\n",
       "      <td>0.432007</td>\n",
       "      <td>-1.16884</td>\n",
       "      <td>0.419284</td>\n",
       "      <td>-1.334440</td>\n",
       "      <td>-2.821480</td>\n",
       "      <td>1.15786</td>\n",
       "      <td>2.676690</td>\n",
       "      <td>-0.005938</td>\n",
       "      <td>0.038004</td>\n",
       "      <td>...</td>\n",
       "      <td>1.137990</td>\n",
       "      <td>0.594746</td>\n",
       "      <td>2.45122</td>\n",
       "      <td>0.424864</td>\n",
       "      <td>1.923710</td>\n",
       "      <td>1.07483</td>\n",
       "      <td>0.986610</td>\n",
       "      <td>0.093465</td>\n",
       "      <td>0.045241</td>\n",
       "      <td>2.145100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dmr_316</th>\n",
       "      <td>7.50020</td>\n",
       "      <td>-0.016628</td>\n",
       "      <td>1.01214</td>\n",
       "      <td>14.927100</td>\n",
       "      <td>4.129020</td>\n",
       "      <td>1.911680</td>\n",
       "      <td>13.18620</td>\n",
       "      <td>1.360270</td>\n",
       "      <td>8.352440</td>\n",
       "      <td>0.329248</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.084340</td>\n",
       "      <td>6.393690</td>\n",
       "      <td>2.88845</td>\n",
       "      <td>25.426200</td>\n",
       "      <td>22.933700</td>\n",
       "      <td>15.93210</td>\n",
       "      <td>11.058300</td>\n",
       "      <td>6.626770</td>\n",
       "      <td>12.788000</td>\n",
       "      <td>2.838900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00092197</th>\n",
       "      <td>-11.77950</td>\n",
       "      <td>-2.961070</td>\n",
       "      <td>-6.99260</td>\n",
       "      <td>-7.504920</td>\n",
       "      <td>-4.824750</td>\n",
       "      <td>-3.665080</td>\n",
       "      <td>-2.74892</td>\n",
       "      <td>-4.332460</td>\n",
       "      <td>-4.197160</td>\n",
       "      <td>-2.492510</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.170910</td>\n",
       "      <td>-11.198800</td>\n",
       "      <td>-4.76573</td>\n",
       "      <td>-8.996460</td>\n",
       "      <td>-11.898900</td>\n",
       "      <td>-6.75238</td>\n",
       "      <td>-8.517520</td>\n",
       "      <td>-3.350240</td>\n",
       "      <td>-5.480380</td>\n",
       "      <td>-3.448570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00106057</th>\n",
       "      <td>-22.22970</td>\n",
       "      <td>-8.121470</td>\n",
       "      <td>-9.95687</td>\n",
       "      <td>-9.213080</td>\n",
       "      <td>-1.399430</td>\n",
       "      <td>-5.819780</td>\n",
       "      <td>-3.07508</td>\n",
       "      <td>-2.980690</td>\n",
       "      <td>-4.411350</td>\n",
       "      <td>-4.824840</td>\n",
       "      <td>...</td>\n",
       "      <td>-8.517060</td>\n",
       "      <td>-12.004400</td>\n",
       "      <td>-7.94205</td>\n",
       "      <td>-8.859080</td>\n",
       "      <td>-16.778100</td>\n",
       "      <td>-18.80480</td>\n",
       "      <td>-13.504700</td>\n",
       "      <td>-15.415900</td>\n",
       "      <td>-12.599400</td>\n",
       "      <td>-9.197450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00122731</th>\n",
       "      <td>-18.58320</td>\n",
       "      <td>-7.004070</td>\n",
       "      <td>-10.09400</td>\n",
       "      <td>-8.986480</td>\n",
       "      <td>-3.159230</td>\n",
       "      <td>-5.798770</td>\n",
       "      <td>-3.49107</td>\n",
       "      <td>-6.444480</td>\n",
       "      <td>-9.071440</td>\n",
       "      <td>-9.405050</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.791460</td>\n",
       "      <td>-12.514200</td>\n",
       "      <td>-7.09567</td>\n",
       "      <td>-10.256600</td>\n",
       "      <td>-14.578500</td>\n",
       "      <td>-16.57800</td>\n",
       "      <td>-13.171400</td>\n",
       "      <td>-12.330000</td>\n",
       "      <td>-14.361300</td>\n",
       "      <td>-9.111960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NegativeControl</th>\n",
       "      <td>-166.04000</td>\n",
       "      <td>-118.832000</td>\n",
       "      <td>-130.24200</td>\n",
       "      <td>-124.836000</td>\n",
       "      <td>-107.220000</td>\n",
       "      <td>-124.721000</td>\n",
       "      <td>-111.83800</td>\n",
       "      <td>-113.765000</td>\n",
       "      <td>-116.104000</td>\n",
       "      <td>-110.328000</td>\n",
       "      <td>...</td>\n",
       "      <td>-103.125000</td>\n",
       "      <td>-127.329000</td>\n",
       "      <td>-98.16780</td>\n",
       "      <td>-124.172000</td>\n",
       "      <td>-141.184000</td>\n",
       "      <td>-137.20100</td>\n",
       "      <td>-128.727000</td>\n",
       "      <td>-134.198000</td>\n",
       "      <td>-123.992000</td>\n",
       "      <td>-112.417000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCorner3</th>\n",
       "      <td>1.76472</td>\n",
       "      <td>0.763844</td>\n",
       "      <td>0.49133</td>\n",
       "      <td>2.206300</td>\n",
       "      <td>0.246871</td>\n",
       "      <td>0.493364</td>\n",
       "      <td>1.16855</td>\n",
       "      <td>1.135890</td>\n",
       "      <td>1.866210</td>\n",
       "      <td>2.006750</td>\n",
       "      <td>...</td>\n",
       "      <td>1.910350</td>\n",
       "      <td>3.086510</td>\n",
       "      <td>2.73111</td>\n",
       "      <td>0.457707</td>\n",
       "      <td>0.286894</td>\n",
       "      <td>1.48711</td>\n",
       "      <td>1.179430</td>\n",
       "      <td>0.984978</td>\n",
       "      <td>0.427812</td>\n",
       "      <td>-0.398997</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>821 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM495879   GSM495880  GSM495881   GSM495882   GSM495883  \\\n",
       "ID                                                                          \n",
       "DarkCorner        -7.21640   -6.898350   -9.59862   -4.718430   -9.114260   \n",
       "dmr_285           -3.24370    0.029345   -2.96027   -3.947260   -2.508740   \n",
       "dmr_3             -4.14763   -2.179640   -1.86509   -2.264490   -1.913330   \n",
       "dmr_308           -2.56221    0.432007   -1.16884    0.419284   -1.334440   \n",
       "dmr_316            7.50020   -0.016628    1.01214   14.927100    4.129020   \n",
       "...                    ...         ...        ...         ...         ...   \n",
       "NC2_00092197     -11.77950   -2.961070   -6.99260   -7.504920   -4.824750   \n",
       "NC2_00106057     -22.22970   -8.121470   -9.95687   -9.213080   -1.399430   \n",
       "NC2_00122731     -18.58320   -7.004070  -10.09400   -8.986480   -3.159230   \n",
       "NegativeControl -166.04000 -118.832000 -130.24200 -124.836000 -107.220000   \n",
       "SCorner3           1.76472    0.763844    0.49133    2.206300    0.246871   \n",
       "\n",
       "                  GSM495884  GSM495885   GSM495886   GSM495887   GSM495888  \\\n",
       "ID                                                                           \n",
       "DarkCorner        -5.497120   -8.27575   -0.418281   -7.752440    5.263860   \n",
       "dmr_285           -2.449810   -2.76578   -2.222550   -2.485020    0.194956   \n",
       "dmr_3             -3.938300   -2.03988   -0.926865    1.122930   -0.061020   \n",
       "dmr_308           -2.821480    1.15786    2.676690   -0.005938    0.038004   \n",
       "dmr_316            1.911680   13.18620    1.360270    8.352440    0.329248   \n",
       "...                     ...        ...         ...         ...         ...   \n",
       "NC2_00092197      -3.665080   -2.74892   -4.332460   -4.197160   -2.492510   \n",
       "NC2_00106057      -5.819780   -3.07508   -2.980690   -4.411350   -4.824840   \n",
       "NC2_00122731      -5.798770   -3.49107   -6.444480   -9.071440   -9.405050   \n",
       "NegativeControl -124.721000 -111.83800 -113.765000 -116.104000 -110.328000   \n",
       "SCorner3           0.493364    1.16855    1.135890    1.866210    2.006750   \n",
       "\n",
       "                 ...   GSM495899   GSM495900  GSM495901   GSM495902  \\\n",
       "ID               ...                                                  \n",
       "DarkCorner       ...   -6.673030  -10.034100   -2.87338   -6.890780   \n",
       "dmr_285          ...   -0.622680   -3.282510   -1.49847   -3.208890   \n",
       "dmr_3            ...    0.368513    1.808110    2.45639    0.092283   \n",
       "dmr_308          ...    1.137990    0.594746    2.45122    0.424864   \n",
       "dmr_316          ...   -1.084340    6.393690    2.88845   25.426200   \n",
       "...              ...         ...         ...        ...         ...   \n",
       "NC2_00092197     ...   -6.170910  -11.198800   -4.76573   -8.996460   \n",
       "NC2_00106057     ...   -8.517060  -12.004400   -7.94205   -8.859080   \n",
       "NC2_00122731     ...   -7.791460  -12.514200   -7.09567  -10.256600   \n",
       "NegativeControl  ... -103.125000 -127.329000  -98.16780 -124.172000   \n",
       "SCorner3         ...    1.910350    3.086510    2.73111    0.457707   \n",
       "\n",
       "                  GSM495903  GSM495904   GSM495905   GSM495906   GSM495907  \\\n",
       "ID                                                                           \n",
       "DarkCorner       -13.643800  -11.24230  -14.145600  -13.803400  -11.586800   \n",
       "dmr_285           -1.656840   -2.42992   -2.638710   -0.923110   -1.950690   \n",
       "dmr_3             -0.197860    1.94856    0.776828    4.097750    0.194788   \n",
       "dmr_308            1.923710    1.07483    0.986610    0.093465    0.045241   \n",
       "dmr_316           22.933700   15.93210   11.058300    6.626770   12.788000   \n",
       "...                     ...        ...         ...         ...         ...   \n",
       "NC2_00092197     -11.898900   -6.75238   -8.517520   -3.350240   -5.480380   \n",
       "NC2_00106057     -16.778100  -18.80480  -13.504700  -15.415900  -12.599400   \n",
       "NC2_00122731     -14.578500  -16.57800  -13.171400  -12.330000  -14.361300   \n",
       "NegativeControl -141.184000 -137.20100 -128.727000 -134.198000 -123.992000   \n",
       "SCorner3           0.286894    1.48711    1.179430    0.984978    0.427812   \n",
       "\n",
       "                  GSM495908  \n",
       "ID                           \n",
       "DarkCorner       -11.498000  \n",
       "dmr_285            1.098530  \n",
       "dmr_3              2.004070  \n",
       "dmr_308            2.145100  \n",
       "dmr_316            2.838900  \n",
       "...                     ...  \n",
       "NC2_00092197      -3.448570  \n",
       "NC2_00106057      -9.197450  \n",
       "NC2_00122731      -9.111960  \n",
       "NegativeControl -112.417000  \n",
       "SCorner3          -0.398997  \n",
       "\n",
       "[821 rows x 30 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['hsa-let-7a', 'hsa-let-7a*', 'hsa-let-7b', 'hsa-let-7b*', 'hsa-let-7c'], 'ControlType': ['0', '0', '0', '0', '0'], 'ORGANISM': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'miRNA_ID': ['hsa-let-7a', 'hsa-let-7a*', 'hsa-let-7b', 'hsa-let-7b*', 'hsa-let-7c'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'SPOT_ID.1': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['DarkCorner', 'dmr_285', 'dmr_3', 'dmr_308', 'dmr_316', 'dmr_31a', 'dmr_6', 'ebv-miR-BART1-3p', 'ebv-miR-BART1-5p', 'ebv-miR-BART10', 'ebv-miR-BART10*', 'ebv-miR-BART11-3p', 'ebv-miR-BART11-5p', 'ebv-miR-BART12', 'ebv-miR-BART13', 'ebv-miR-BART13*', 'ebv-miR-BART14', 'ebv-miR-BART14*', 'ebv-miR-BART15', 'ebv-miR-BART16']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['hsa-let-7a', 'hsa-let-7a*', 'hsa-let-7b', 'hsa-let-7b*', 'hsa-let-7c'], 'ControlType': ['0', '0', '0', '0', '0'], 'ORGANISM': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'miRNA_ID': ['hsa-let-7a', 'hsa-let-7a*', 'hsa-let-7b', 'hsa-let-7b*', 'hsa-let-7c'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'SPOT_ID.1': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ControlType</th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>SPOT_ID.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a*</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7a*</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b*</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7b*</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7c</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>hsa-let-7c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25536</th>\n",
       "      <td>kshv-miR-K12-8</td>\n",
       "      <td>26.114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25537</th>\n",
       "      <td>kshv-miR-K12-9</td>\n",
       "      <td>-0.642957</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25538</th>\n",
       "      <td>kshv-miR-K12-9*</td>\n",
       "      <td>0.699658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25539</th>\n",
       "      <td>miRNABrightCorner30</td>\n",
       "      <td>106091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25540</th>\n",
       "      <td>mr_1</td>\n",
       "      <td>5913.86</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25541 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        ID ControlType      ORGANISM     miRNA_ID SPOT_ID  \\\n",
       "0               hsa-let-7a           0  Homo sapiens   hsa-let-7a     NaN   \n",
       "1              hsa-let-7a*           0  Homo sapiens  hsa-let-7a*     NaN   \n",
       "2               hsa-let-7b           0  Homo sapiens   hsa-let-7b     NaN   \n",
       "3              hsa-let-7b*           0  Homo sapiens  hsa-let-7b*     NaN   \n",
       "4               hsa-let-7c           0  Homo sapiens   hsa-let-7c     NaN   \n",
       "...                    ...         ...           ...          ...     ...   \n",
       "25536       kshv-miR-K12-8      26.114           NaN          NaN     NaN   \n",
       "25537       kshv-miR-K12-9   -0.642957           NaN          NaN     NaN   \n",
       "25538      kshv-miR-K12-9*    0.699658           NaN          NaN     NaN   \n",
       "25539  miRNABrightCorner30      106091           NaN          NaN     NaN   \n",
       "25540                 mr_1     5913.86           NaN          NaN     NaN   \n",
       "\n",
       "       SPOT_ID.1  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "...          ...  \n",
       "25536        NaN  \n",
       "25537        NaN  \n",
       "25538        NaN  \n",
       "25539        NaN  \n",
       "25540        NaN  \n",
       "\n",
       "[25541 rows x 6 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'ControlType', 'ORGANISM', 'miRNA_ID', 'SPOT_ID', 'SPOT_ID.1'], dtype='object')"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Identity by descent mapping of founder mutations in cancer using high-resolution tumor SNP data\"\n",
      "!Series_summary\t\"We present a computational tool, FounderTracker, for discovering founder mutations in cancer, based on the detection of significantly conserved haplotypes in tumor SNP profiles. We demonstrate the relevance of the approach by identifying founder mutations in two different cancers, and we show with simulated data that FounderTracker can detect rare founder mutations with high power and negligible false discovery rate. FounderTracker is a powerful tool for discovering novel founder mutations that may explain part of the \"\"missing\"\" heritability in cancer.\"\n",
      "!Series_summary\t\"\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM610697</th>\n",
       "      <th>GSM610698</th>\n",
       "      <th>GSM610699</th>\n",
       "      <th>GSM610700</th>\n",
       "      <th>GSM610701</th>\n",
       "      <th>GSM610702</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "      <td>geographical origin: Brazil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Male</td>\n",
       "      <td>genger: Female</td>\n",
       "      <td>genger: Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age  (months): 42.96</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>histological type: Paired blood sample corresp...</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tnm: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tnm: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tp53 mutation: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>virilization: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>virilization: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cushing  syndrome: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>treatment: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>treatment: --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>follow-up (months): --</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "5  !Sample_characteristics_ch1   \n",
       "6  !Sample_characteristics_ch1   \n",
       "7  !Sample_characteristics_ch1   \n",
       "8  !Sample_characteristics_ch1   \n",
       "9  !Sample_characteristics_ch1   \n",
       "\n",
       "                                           GSM610697  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2                               age  (months): 42.96   \n",
       "3  histological type: Paired blood sample corresp...   \n",
       "4                                            tnm: --   \n",
       "5                                  tp53 mutation: --   \n",
       "6                                   virilization: --   \n",
       "7                              cushing  syndrome: --   \n",
       "8                                      treatment: --   \n",
       "9                             follow-up (months): --   \n",
       "\n",
       "                                           GSM610698  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610699  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610700  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                       genger: Male   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610701  \\\n",
       "0                        geographical origin: Brazil   \n",
       "1                                     genger: Female   \n",
       "2  histological type: Paired blood sample corresp...   \n",
       "3                                            tnm: --   \n",
       "4                                  tp53 mutation: --   \n",
       "5                                   virilization: --   \n",
       "6                              cushing  syndrome: --   \n",
       "7                                      treatment: --   \n",
       "8                             follow-up (months): --   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM610702  \n",
       "0                        geographical origin: Brazil  \n",
       "1                                     genger: Female  \n",
       "2  histological type: Paired blood sample corresp...  \n",
       "3                                            tnm: --  \n",
       "4                                  tp53 mutation: --  \n",
       "5                                   virilization: --  \n",
       "6                              cushing  syndrome: --  \n",
       "7                                      treatment: --  \n",
       "8                             follow-up (months): --  \n",
       "9                                                NaN  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stop: NO trait\n",
    "\n",
    "cohort = accession_num = \"GSE32206\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene exrpression profiling of childhood adrenocortical tumors\"\n",
      "!Series_summary\t\"Pediatric adrenocortical tumors (ACT) are rare and often fatal malignancies; little is known regarding their etiology and biology. To provide additional insight into the nature of ACT, we determined the gene expression profiles of 24 pediatric tumors (five adenomas, 18 carcinomas, and one undetermined) and seven normal adrenal glands. Distinct patterns of gene expression, validated by quantitative real-time PCR and Western blot analysis, were identified that distinguish normal adrenal cortex from tumor. Differences in gene expression were also identified between adrenocortical adenomas and carcinomas. In addition, pediatric adrenocortical carcinomas were found to share similar patterns of gene expression when compared with those published for adult ACT. This study represents the first microarray analysis of childhood ACT. Our findings lay the groundwork for establishing gene expression profiles that may aid in the diagnosis and prognosis of pediatric ACT, and in the identification of signaling pathways that contribute to this disease.\"\n",
      "!Series_overall_design\t\"We used microarrays to explore the expression profiles differentially expressed in childhood adrenocortical tumors and in normal adrenal gland tissues. Pediatric adrenocortical adenoma and carcinoma patients were enrolled on the International Pediatric Adrenocortical Tumor Registry (IPACTR) and Bank protocol. Tumor specimens were harvested during surgery and snap frozen in liquid nitrogen to preserve tissue integrity. Data have been compiled for eight males and 15 females between 0 and 16 years of age. Table 1 (West et al, Cancer Research 67:601-608, 2007) summarizes the primary clinical information for each subject (excluding sample Unk1 with ACT of undetermined histology), including stage of the disease, tumor class, sex, age, relapse-free survival, and overall survival.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1954726</th>\n",
       "      <th>GSM1954727</th>\n",
       "      <th>GSM1954728</th>\n",
       "      <th>GSM1954729</th>\n",
       "      <th>GSM1954730</th>\n",
       "      <th>GSM1954731</th>\n",
       "      <th>GSM1954732</th>\n",
       "      <th>GSM1954733</th>\n",
       "      <th>GSM1954734</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1954747</th>\n",
       "      <th>GSM1954748</th>\n",
       "      <th>GSM1954749</th>\n",
       "      <th>GSM1954750</th>\n",
       "      <th>GSM1954751</th>\n",
       "      <th>GSM1954752</th>\n",
       "      <th>GSM1954753</th>\n",
       "      <th>GSM1954754</th>\n",
       "      <th>GSM1954755</th>\n",
       "      <th>GSM1954756</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "      <td>gender: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>histologic type: adrenocortical adenoma</td>\n",
       "      <td>histologic type: adrenocortical adenoma</td>\n",
       "      <td>histologic type: adrenocortical adenoma</td>\n",
       "      <td>histologic type: adrenocortical adenoma</td>\n",
       "      <td>histologic type: adrenocortical adenoma</td>\n",
       "      <td>histologic type: adrenocortical carcinoma</td>\n",
       "      <td>histologic type: adrenocortical carcinoma</td>\n",
       "      <td>histologic type: adrenocortical carcinoma</td>\n",
       "      <td>histologic type: adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>histologic type: adrenocortical carcinoma</td>\n",
       "      <td>histologic type: adrenocortical carcinoma</td>\n",
       "      <td>histologic type: unknown</td>\n",
       "      <td>histologic type: normal</td>\n",
       "      <td>histologic type: normal</td>\n",
       "      <td>histologic type: normal</td>\n",
       "      <td>histologic type: normal</td>\n",
       "      <td>histologic type: normal</td>\n",
       "      <td>histologic type: normal</td>\n",
       "      <td>histologic type: normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor stage: not staged</td>\n",
       "      <td>tumor stage: not staged</td>\n",
       "      <td>tumor stage: not staged</td>\n",
       "      <td>tumor stage: not staged</td>\n",
       "      <td>tumor stage: not staged</td>\n",
       "      <td>tumor stage: 4</td>\n",
       "      <td>tumor stage: 2</td>\n",
       "      <td>tumor stage: 3</td>\n",
       "      <td>tumor stage: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor stage: 4</td>\n",
       "      <td>tumor stage: 3</td>\n",
       "      <td>tumor stage: unknown</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "      <td>tumor stage: not applicable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                               GSM1954726  \\\n",
       "0  !Sample_characteristics_ch1                           gender: female   \n",
       "1  !Sample_characteristics_ch1  histologic type: adrenocortical adenoma   \n",
       "2  !Sample_characteristics_ch1                  tumor stage: not staged   \n",
       "\n",
       "                                GSM1954727  \\\n",
       "0                           gender: female   \n",
       "1  histologic type: adrenocortical adenoma   \n",
       "2                  tumor stage: not staged   \n",
       "\n",
       "                                GSM1954728  \\\n",
       "0                           gender: female   \n",
       "1  histologic type: adrenocortical adenoma   \n",
       "2                  tumor stage: not staged   \n",
       "\n",
       "                                GSM1954729  \\\n",
       "0                           gender: female   \n",
       "1  histologic type: adrenocortical adenoma   \n",
       "2                  tumor stage: not staged   \n",
       "\n",
       "                                GSM1954730  \\\n",
       "0                           gender: female   \n",
       "1  histologic type: adrenocortical adenoma   \n",
       "2                  tumor stage: not staged   \n",
       "\n",
       "                                  GSM1954731  \\\n",
       "0                               gender: male   \n",
       "1  histologic type: adrenocortical carcinoma   \n",
       "2                             tumor stage: 4   \n",
       "\n",
       "                                  GSM1954732  \\\n",
       "0                             gender: female   \n",
       "1  histologic type: adrenocortical carcinoma   \n",
       "2                             tumor stage: 2   \n",
       "\n",
       "                                  GSM1954733  \\\n",
       "0                               gender: male   \n",
       "1  histologic type: adrenocortical carcinoma   \n",
       "2                             tumor stage: 3   \n",
       "\n",
       "                                  GSM1954734  ...  \\\n",
       "0                             gender: female  ...   \n",
       "1  histologic type: adrenocortical carcinoma  ...   \n",
       "2                             tumor stage: 1  ...   \n",
       "\n",
       "                                  GSM1954747  \\\n",
       "0                             gender: female   \n",
       "1  histologic type: adrenocortical carcinoma   \n",
       "2                             tumor stage: 4   \n",
       "\n",
       "                                  GSM1954748                GSM1954749  \\\n",
       "0                             gender: female           gender: unknown   \n",
       "1  histologic type: adrenocortical carcinoma  histologic type: unknown   \n",
       "2                             tumor stage: 3      tumor stage: unknown   \n",
       "\n",
       "                    GSM1954750                   GSM1954751  \\\n",
       "0              gender: unknown              gender: unknown   \n",
       "1      histologic type: normal      histologic type: normal   \n",
       "2  tumor stage: not applicable  tumor stage: not applicable   \n",
       "\n",
       "                    GSM1954752                   GSM1954753  \\\n",
       "0              gender: unknown              gender: unknown   \n",
       "1      histologic type: normal      histologic type: normal   \n",
       "2  tumor stage: not applicable  tumor stage: not applicable   \n",
       "\n",
       "                    GSM1954754                   GSM1954755  \\\n",
       "0              gender: unknown              gender: unknown   \n",
       "1      histologic type: normal      histologic type: normal   \n",
       "2  tumor stage: not applicable  tumor stage: not applicable   \n",
       "\n",
       "                    GSM1954756  \n",
       "0              gender: unknown  \n",
       "1      histologic type: normal  \n",
       "2  tumor stage: not applicable  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "\n",
    "cohort = accession_num = \"GSE75415\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'histologic type: adrenocortical adenoma',\n",
       "       'histologic type: adrenocortical carcinoma',\n",
       "       'histologic type: unknown', 'histologic type: normal'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'histologic type: normal':\n",
    "        return 0  # Epilepsy present\n",
    "    elif tissue_type == 'histologic type: unknown':\n",
    "        return None  # Epilepsy not present\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1954726</th>\n",
       "      <th>GSM1954727</th>\n",
       "      <th>GSM1954728</th>\n",
       "      <th>GSM1954729</th>\n",
       "      <th>GSM1954730</th>\n",
       "      <th>GSM1954731</th>\n",
       "      <th>GSM1954732</th>\n",
       "      <th>GSM1954733</th>\n",
       "      <th>GSM1954734</th>\n",
       "      <th>GSM1954735</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1954747</th>\n",
       "      <th>GSM1954748</th>\n",
       "      <th>GSM1954749</th>\n",
       "      <th>GSM1954750</th>\n",
       "      <th>GSM1954751</th>\n",
       "      <th>GSM1954752</th>\n",
       "      <th>GSM1954753</th>\n",
       "      <th>GSM1954754</th>\n",
       "      <th>GSM1954755</th>\n",
       "      <th>GSM1954756</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1954726  GSM1954727  GSM1954728  GSM1954729  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Gender                          1           1           1           1   \n",
       "\n",
       "                       GSM1954730  GSM1954731  GSM1954732  GSM1954733  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Gender                          1           0           1           0   \n",
       "\n",
       "                       GSM1954734  GSM1954735  ...  GSM1954747  GSM1954748  \\\n",
       "Adrenocortical Cancer           1           1  ...           1           1   \n",
       "Gender                          1           0  ...           1           1   \n",
       "\n",
       "                       GSM1954749  GSM1954750  GSM1954751  GSM1954752  \\\n",
       "Adrenocortical Cancer        None           0           0           0   \n",
       "Gender                       None        None        None        None   \n",
       "\n",
       "                       GSM1954753  GSM1954754  GSM1954755  GSM1954756  \n",
       "Adrenocortical Cancer           0           0           0           0  \n",
       "Gender                       None        None        None        None  \n",
       "\n",
       "[2 rows x 31 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1954726</th>\n",
       "      <th>GSM1954727</th>\n",
       "      <th>GSM1954728</th>\n",
       "      <th>GSM1954729</th>\n",
       "      <th>GSM1954730</th>\n",
       "      <th>GSM1954731</th>\n",
       "      <th>GSM1954732</th>\n",
       "      <th>GSM1954733</th>\n",
       "      <th>GSM1954734</th>\n",
       "      <th>GSM1954735</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1954747</th>\n",
       "      <th>GSM1954748</th>\n",
       "      <th>GSM1954749</th>\n",
       "      <th>GSM1954750</th>\n",
       "      <th>GSM1954751</th>\n",
       "      <th>GSM1954752</th>\n",
       "      <th>GSM1954753</th>\n",
       "      <th>GSM1954754</th>\n",
       "      <th>GSM1954755</th>\n",
       "      <th>GSM1954756</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>8001.9</td>\n",
       "      <td>2372.1</td>\n",
       "      <td>2968.4</td>\n",
       "      <td>3541.5</td>\n",
       "      <td>1470.0</td>\n",
       "      <td>3469.1</td>\n",
       "      <td>4500.6</td>\n",
       "      <td>1744.5</td>\n",
       "      <td>3630.4</td>\n",
       "      <td>5535.4</td>\n",
       "      <td>...</td>\n",
       "      <td>3939.5</td>\n",
       "      <td>2257.8</td>\n",
       "      <td>3697.5</td>\n",
       "      <td>1613.3</td>\n",
       "      <td>1868.5</td>\n",
       "      <td>2791.8</td>\n",
       "      <td>2650.7</td>\n",
       "      <td>2430.2</td>\n",
       "      <td>2373.9</td>\n",
       "      <td>2853.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>243.7</td>\n",
       "      <td>204.3</td>\n",
       "      <td>42.3</td>\n",
       "      <td>165.6</td>\n",
       "      <td>113.3</td>\n",
       "      <td>221.4</td>\n",
       "      <td>259.2</td>\n",
       "      <td>415.3</td>\n",
       "      <td>192.2</td>\n",
       "      <td>224.9</td>\n",
       "      <td>...</td>\n",
       "      <td>222.0</td>\n",
       "      <td>181.5</td>\n",
       "      <td>127.7</td>\n",
       "      <td>186.1</td>\n",
       "      <td>55.2</td>\n",
       "      <td>82.7</td>\n",
       "      <td>87.5</td>\n",
       "      <td>103.0</td>\n",
       "      <td>67.4</td>\n",
       "      <td>141.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>475.3</td>\n",
       "      <td>255.5</td>\n",
       "      <td>158.8</td>\n",
       "      <td>135.8</td>\n",
       "      <td>134.2</td>\n",
       "      <td>1113.7</td>\n",
       "      <td>392.2</td>\n",
       "      <td>166.6</td>\n",
       "      <td>124.8</td>\n",
       "      <td>227.2</td>\n",
       "      <td>...</td>\n",
       "      <td>281.0</td>\n",
       "      <td>1527.1</td>\n",
       "      <td>170.7</td>\n",
       "      <td>205.8</td>\n",
       "      <td>160.9</td>\n",
       "      <td>331.1</td>\n",
       "      <td>96.5</td>\n",
       "      <td>218.1</td>\n",
       "      <td>314.5</td>\n",
       "      <td>124.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>961.4</td>\n",
       "      <td>1164.1</td>\n",
       "      <td>1469.3</td>\n",
       "      <td>830.7</td>\n",
       "      <td>1145.5</td>\n",
       "      <td>1762.3</td>\n",
       "      <td>1189.0</td>\n",
       "      <td>976.4</td>\n",
       "      <td>808.3</td>\n",
       "      <td>1267.2</td>\n",
       "      <td>...</td>\n",
       "      <td>1751.6</td>\n",
       "      <td>1923.4</td>\n",
       "      <td>946.1</td>\n",
       "      <td>1332.6</td>\n",
       "      <td>1412.8</td>\n",
       "      <td>1611.8</td>\n",
       "      <td>2005.8</td>\n",
       "      <td>1548.6</td>\n",
       "      <td>2192.5</td>\n",
       "      <td>1804.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>83.7</td>\n",
       "      <td>28.6</td>\n",
       "      <td>15.0</td>\n",
       "      <td>67.3</td>\n",
       "      <td>65.4</td>\n",
       "      <td>105.8</td>\n",
       "      <td>168.0</td>\n",
       "      <td>31.1</td>\n",
       "      <td>73.7</td>\n",
       "      <td>158.4</td>\n",
       "      <td>...</td>\n",
       "      <td>122.4</td>\n",
       "      <td>176.8</td>\n",
       "      <td>77.4</td>\n",
       "      <td>96.1</td>\n",
       "      <td>192.5</td>\n",
       "      <td>14.6</td>\n",
       "      <td>120.6</td>\n",
       "      <td>311.7</td>\n",
       "      <td>225.9</td>\n",
       "      <td>94.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>10.1</td>\n",
       "      <td>15.7</td>\n",
       "      <td>71.6</td>\n",
       "      <td>17.3</td>\n",
       "      <td>29.5</td>\n",
       "      <td>11.9</td>\n",
       "      <td>23.7</td>\n",
       "      <td>13.8</td>\n",
       "      <td>11.3</td>\n",
       "      <td>72.7</td>\n",
       "      <td>...</td>\n",
       "      <td>10.5</td>\n",
       "      <td>30.8</td>\n",
       "      <td>18.1</td>\n",
       "      <td>8.5</td>\n",
       "      <td>170.8</td>\n",
       "      <td>42.4</td>\n",
       "      <td>31.5</td>\n",
       "      <td>26.3</td>\n",
       "      <td>18.3</td>\n",
       "      <td>34.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>146.3</td>\n",
       "      <td>95.0</td>\n",
       "      <td>80.4</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>37.5</td>\n",
       "      <td>176.1</td>\n",
       "      <td>5.2</td>\n",
       "      <td>5.2</td>\n",
       "      <td>94.8</td>\n",
       "      <td>...</td>\n",
       "      <td>130.0</td>\n",
       "      <td>37.4</td>\n",
       "      <td>56.3</td>\n",
       "      <td>64.2</td>\n",
       "      <td>28.4</td>\n",
       "      <td>62.1</td>\n",
       "      <td>73.8</td>\n",
       "      <td>29.1</td>\n",
       "      <td>79.0</td>\n",
       "      <td>56.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>118.6</td>\n",
       "      <td>13.2</td>\n",
       "      <td>10.2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>52.6</td>\n",
       "      <td>25.7</td>\n",
       "      <td>58.0</td>\n",
       "      <td>3.7</td>\n",
       "      <td>3.4</td>\n",
       "      <td>45.4</td>\n",
       "      <td>...</td>\n",
       "      <td>120.3</td>\n",
       "      <td>38.3</td>\n",
       "      <td>40.5</td>\n",
       "      <td>2.9</td>\n",
       "      <td>10.3</td>\n",
       "      <td>8.2</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.3</td>\n",
       "      <td>58.3</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>73.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>63.3</td>\n",
       "      <td>5.9</td>\n",
       "      <td>27.4</td>\n",
       "      <td>11.8</td>\n",
       "      <td>74.6</td>\n",
       "      <td>10.1</td>\n",
       "      <td>10.3</td>\n",
       "      <td>75.1</td>\n",
       "      <td>...</td>\n",
       "      <td>20.4</td>\n",
       "      <td>108.6</td>\n",
       "      <td>59.1</td>\n",
       "      <td>4.5</td>\n",
       "      <td>26.4</td>\n",
       "      <td>19.5</td>\n",
       "      <td>11.2</td>\n",
       "      <td>65.8</td>\n",
       "      <td>38.3</td>\n",
       "      <td>129.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>53.6</td>\n",
       "      <td>56.0</td>\n",
       "      <td>3.9</td>\n",
       "      <td>7.9</td>\n",
       "      <td>2.7</td>\n",
       "      <td>11.7</td>\n",
       "      <td>20.1</td>\n",
       "      <td>28.3</td>\n",
       "      <td>7.5</td>\n",
       "      <td>4.5</td>\n",
       "      <td>...</td>\n",
       "      <td>6.4</td>\n",
       "      <td>21.4</td>\n",
       "      <td>11.9</td>\n",
       "      <td>6.0</td>\n",
       "      <td>12.8</td>\n",
       "      <td>12.5</td>\n",
       "      <td>9.9</td>\n",
       "      <td>30.3</td>\n",
       "      <td>7.5</td>\n",
       "      <td>16.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22283 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM1954726  GSM1954727  GSM1954728  GSM1954729  GSM1954730  \\\n",
       "ID                                                                            \n",
       "1007_s_at            8001.9      2372.1      2968.4      3541.5      1470.0   \n",
       "1053_at               243.7       204.3        42.3       165.6       113.3   \n",
       "117_at                475.3       255.5       158.8       135.8       134.2   \n",
       "121_at                961.4      1164.1      1469.3       830.7      1145.5   \n",
       "1255_g_at              83.7        28.6        15.0        67.3        65.4   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at         10.1        15.7        71.6        17.3        29.5   \n",
       "AFFX-ThrX-M_at        146.3        95.0        80.4        13.0         2.3   \n",
       "AFFX-TrpnX-3_at       118.6        13.2        10.2         6.0        52.6   \n",
       "AFFX-TrpnX-5_at        73.0        18.0        63.3         5.9        27.4   \n",
       "AFFX-TrpnX-M_at        53.6        56.0         3.9         7.9         2.7   \n",
       "\n",
       "                 GSM1954731  GSM1954732  GSM1954733  GSM1954734  GSM1954735  \\\n",
       "ID                                                                            \n",
       "1007_s_at            3469.1      4500.6      1744.5      3630.4      5535.4   \n",
       "1053_at               221.4       259.2       415.3       192.2       224.9   \n",
       "117_at               1113.7       392.2       166.6       124.8       227.2   \n",
       "121_at               1762.3      1189.0       976.4       808.3      1267.2   \n",
       "1255_g_at             105.8       168.0        31.1        73.7       158.4   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at         11.9        23.7        13.8        11.3        72.7   \n",
       "AFFX-ThrX-M_at         37.5       176.1         5.2         5.2        94.8   \n",
       "AFFX-TrpnX-3_at        25.7        58.0         3.7         3.4        45.4   \n",
       "AFFX-TrpnX-5_at        11.8        74.6        10.1        10.3        75.1   \n",
       "AFFX-TrpnX-M_at        11.7        20.1        28.3         7.5         4.5   \n",
       "\n",
       "                 ...  GSM1954747  GSM1954748  GSM1954749  GSM1954750  \\\n",
       "ID               ...                                                   \n",
       "1007_s_at        ...      3939.5      2257.8      3697.5      1613.3   \n",
       "1053_at          ...       222.0       181.5       127.7       186.1   \n",
       "117_at           ...       281.0      1527.1       170.7       205.8   \n",
       "121_at           ...      1751.6      1923.4       946.1      1332.6   \n",
       "1255_g_at        ...       122.4       176.8        77.4        96.1   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at   ...        10.5        30.8        18.1         8.5   \n",
       "AFFX-ThrX-M_at   ...       130.0        37.4        56.3        64.2   \n",
       "AFFX-TrpnX-3_at  ...       120.3        38.3        40.5         2.9   \n",
       "AFFX-TrpnX-5_at  ...        20.4       108.6        59.1         4.5   \n",
       "AFFX-TrpnX-M_at  ...         6.4        21.4        11.9         6.0   \n",
       "\n",
       "                 GSM1954751  GSM1954752  GSM1954753  GSM1954754  GSM1954755  \\\n",
       "ID                                                                            \n",
       "1007_s_at            1868.5      2791.8      2650.7      2430.2      2373.9   \n",
       "1053_at                55.2        82.7        87.5       103.0        67.4   \n",
       "117_at                160.9       331.1        96.5       218.1       314.5   \n",
       "121_at               1412.8      1611.8      2005.8      1548.6      2192.5   \n",
       "1255_g_at             192.5        14.6       120.6       311.7       225.9   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at        170.8        42.4        31.5        26.3        18.3   \n",
       "AFFX-ThrX-M_at         28.4        62.1        73.8        29.1        79.0   \n",
       "AFFX-TrpnX-3_at        10.3         8.2         5.0         9.3        58.3   \n",
       "AFFX-TrpnX-5_at        26.4        19.5        11.2        65.8        38.3   \n",
       "AFFX-TrpnX-M_at        12.8        12.5         9.9        30.3         7.5   \n",
       "\n",
       "                 GSM1954756  \n",
       "ID                           \n",
       "1007_s_at            2853.3  \n",
       "1053_at               141.4  \n",
       "117_at                124.1  \n",
       "121_at               1804.3  \n",
       "1255_g_at              94.9  \n",
       "...                     ...  \n",
       "AFFX-ThrX-5_at         34.9  \n",
       "AFFX-ThrX-M_at         56.5  \n",
       "AFFX-TrpnX-3_at         5.0  \n",
       "AFFX-TrpnX-5_at       129.2  \n",
       "AFFX-TrpnX-M_at        16.5  \n",
       "\n",
       "[22283 rows x 31 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1598_g_at', '160020_at', '1729_at', '1773_at', '177_at', '179_at', '1861_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713082</th>\n",
       "      <td>222380_s_at</td>\n",
       "      <td>336.9</td>\n",
       "      <td>P</td>\n",
       "      <td>0.008057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713083</th>\n",
       "      <td>222381_at</td>\n",
       "      <td>79.8</td>\n",
       "      <td>A</td>\n",
       "      <td>0.466064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713084</th>\n",
       "      <td>222382_x_at</td>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>0.696289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713085</th>\n",
       "      <td>222383_s_at</td>\n",
       "      <td>25.6</td>\n",
       "      <td>A</td>\n",
       "      <td>0.870361</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713086</th>\n",
       "      <td>222384_at</td>\n",
       "      <td>192.5</td>\n",
       "      <td>A</td>\n",
       "      <td>0.27417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>713087 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID  GB_ACC SPOT_ID Species Scientific Name Annotation Date  \\\n",
       "0         1007_s_at  U48705     NaN            Homo sapiens     Oct 6, 2014   \n",
       "1           1053_at  M87338     NaN            Homo sapiens     Oct 6, 2014   \n",
       "2            117_at  X51757     NaN            Homo sapiens     Oct 6, 2014   \n",
       "3            121_at  X69699     NaN            Homo sapiens     Oct 6, 2014   \n",
       "4         1255_g_at  L36861     NaN            Homo sapiens     Oct 6, 2014   \n",
       "...             ...     ...     ...                     ...             ...   \n",
       "713082  222380_s_at   336.9       P                0.008057             NaN   \n",
       "713083    222381_at    79.8       A                0.466064             NaN   \n",
       "713084  222382_x_at      10       A                0.696289             NaN   \n",
       "713085  222383_s_at    25.6       A                0.870361             NaN   \n",
       "713086    222384_at   192.5       A                 0.27417             NaN   \n",
       "\n",
       "            Sequence Type                  Sequence Source  \\\n",
       "0       Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1       Exemplar sequence                          GenBank   \n",
       "2       Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3       Exemplar sequence                          GenBank   \n",
       "4       Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                   ...                              ...   \n",
       "713082                NaN                              NaN   \n",
       "713083                NaN                              NaN   \n",
       "713084                NaN                              NaN   \n",
       "713085                NaN                              NaN   \n",
       "713086                NaN                              NaN   \n",
       "\n",
       "                                       Target Description  \\\n",
       "0       U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1       M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2       X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3       X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4       L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                   ...   \n",
       "713082                                                NaN   \n",
       "713083                                                NaN   \n",
       "713084                                                NaN   \n",
       "713085                                                NaN   \n",
       "713086                                                NaN   \n",
       "\n",
       "       Representative Public ID  \\\n",
       "0                        U48705   \n",
       "1                        M87338   \n",
       "2                        X51757   \n",
       "3                        X69699   \n",
       "4                        L36861   \n",
       "...                         ...   \n",
       "713082                      NaN   \n",
       "713083                      NaN   \n",
       "713084                      NaN   \n",
       "713085                      NaN   \n",
       "713086                      NaN   \n",
       "\n",
       "                                               Gene Title       Gene Symbol  \\\n",
       "0       discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1             replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2                    heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                            paired box 8              PAX8   \n",
       "4                 guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "...                                                   ...               ...   \n",
       "713082                                                NaN               NaN   \n",
       "713083                                                NaN               NaN   \n",
       "713084                                                NaN               NaN   \n",
       "713085                                                NaN               NaN   \n",
       "713086                                                NaN               NaN   \n",
       "\n",
       "           ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0       780 /// 100616237  NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1                    5982  NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2                    3310                                          NM_002155   \n",
       "3                    7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                    2978                         NM_000409 /// XM_006715073   \n",
       "...                   ...                                                ...   \n",
       "713082                NaN                                                NaN   \n",
       "713083                NaN                                                NaN   \n",
       "713084                NaN                                                NaN   \n",
       "713085                NaN                                                NaN   \n",
       "713086                NaN                                                NaN   \n",
       "\n",
       "                         Gene Ontology Biological Process  \\\n",
       "0       0001558 // regulation of cell growth // inferr...   \n",
       "1       0000278 // mitotic cell cycle // traceable aut...   \n",
       "2       0000902 // cell morphogenesis // inferred from...   \n",
       "3       0001655 // urogenital system development // in...   \n",
       "4       0007165 // signal transduction // non-traceabl...   \n",
       "...                                                   ...   \n",
       "713082                                                NaN   \n",
       "713083                                                NaN   \n",
       "713084                                                NaN   \n",
       "713085                                                NaN   \n",
       "713086                                                NaN   \n",
       "\n",
       "                         Gene Ontology Cellular Component  \\\n",
       "0       0005576 // extracellular region // inferred fr...   \n",
       "1       0005634 // nucleus // inferred from electronic...   \n",
       "2       0005737 // cytoplasm // inferred from direct a...   \n",
       "3       0005634 // nucleus // inferred from direct ass...   \n",
       "4       0001750 // photoreceptor outer segment // infe...   \n",
       "...                                                   ...   \n",
       "713082                                                NaN   \n",
       "713083                                                NaN   \n",
       "713084                                                NaN   \n",
       "713085                                                NaN   \n",
       "713086                                                NaN   \n",
       "\n",
       "                         Gene Ontology Molecular Function  \n",
       "0       0000166 // nucleotide binding // inferred from...  \n",
       "1       0000166 // nucleotide binding // inferred from...  \n",
       "2       0000166 // nucleotide binding // inferred from...  \n",
       "3       0000979 // RNA polymerase II core promoter seq...  \n",
       "4       0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                   ...  \n",
       "713082                                                NaN  \n",
       "713083                                                NaN  \n",
       "713084                                                NaN  \n",
       "713085                                                NaN  \n",
       "713086                                                NaN  \n",
       "\n",
       "[713087 rows x 16 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1954726</th>\n",
       "      <th>GSM1954727</th>\n",
       "      <th>GSM1954728</th>\n",
       "      <th>GSM1954729</th>\n",
       "      <th>GSM1954730</th>\n",
       "      <th>GSM1954731</th>\n",
       "      <th>GSM1954732</th>\n",
       "      <th>GSM1954733</th>\n",
       "      <th>GSM1954734</th>\n",
       "      <th>GSM1954735</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1954747</th>\n",
       "      <th>GSM1954748</th>\n",
       "      <th>GSM1954749</th>\n",
       "      <th>GSM1954750</th>\n",
       "      <th>GSM1954751</th>\n",
       "      <th>GSM1954752</th>\n",
       "      <th>GSM1954753</th>\n",
       "      <th>GSM1954754</th>\n",
       "      <th>GSM1954755</th>\n",
       "      <th>GSM1954756</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>147.500000</td>\n",
       "      <td>695.900000</td>\n",
       "      <td>466.20</td>\n",
       "      <td>1479.40</td>\n",
       "      <td>1155.70</td>\n",
       "      <td>676.0</td>\n",
       "      <td>1514.800000</td>\n",
       "      <td>35.9</td>\n",
       "      <td>202.900000</td>\n",
       "      <td>993.100000</td>\n",
       "      <td>...</td>\n",
       "      <td>405.800000</td>\n",
       "      <td>1016.00</td>\n",
       "      <td>141.5</td>\n",
       "      <td>6708.400000</td>\n",
       "      <td>1930.100000</td>\n",
       "      <td>1995.0</td>\n",
       "      <td>2711.300000</td>\n",
       "      <td>3506.00</td>\n",
       "      <td>4596.500000</td>\n",
       "      <td>2504.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>957.900000</td>\n",
       "      <td>384.200000</td>\n",
       "      <td>320.80</td>\n",
       "      <td>218.10</td>\n",
       "      <td>445.40</td>\n",
       "      <td>238.3</td>\n",
       "      <td>387.700000</td>\n",
       "      <td>495.9</td>\n",
       "      <td>154.200000</td>\n",
       "      <td>241.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>67.700000</td>\n",
       "      <td>245.70</td>\n",
       "      <td>308.5</td>\n",
       "      <td>118.100000</td>\n",
       "      <td>424.400000</td>\n",
       "      <td>235.7</td>\n",
       "      <td>494.700000</td>\n",
       "      <td>563.50</td>\n",
       "      <td>501.000000</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>957.900000</td>\n",
       "      <td>384.200000</td>\n",
       "      <td>320.80</td>\n",
       "      <td>218.10</td>\n",
       "      <td>445.40</td>\n",
       "      <td>238.3</td>\n",
       "      <td>387.700000</td>\n",
       "      <td>495.9</td>\n",
       "      <td>154.200000</td>\n",
       "      <td>241.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>67.700000</td>\n",
       "      <td>245.70</td>\n",
       "      <td>308.5</td>\n",
       "      <td>118.100000</td>\n",
       "      <td>424.400000</td>\n",
       "      <td>235.7</td>\n",
       "      <td>494.700000</td>\n",
       "      <td>563.50</td>\n",
       "      <td>501.000000</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1P2</th>\n",
       "      <td>171.600000</td>\n",
       "      <td>30.100000</td>\n",
       "      <td>26.20</td>\n",
       "      <td>38.70</td>\n",
       "      <td>34.00</td>\n",
       "      <td>34.8</td>\n",
       "      <td>59.300000</td>\n",
       "      <td>82.4</td>\n",
       "      <td>30.200000</td>\n",
       "      <td>176.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>38.700000</td>\n",
       "      <td>78.70</td>\n",
       "      <td>36.6</td>\n",
       "      <td>197.700000</td>\n",
       "      <td>41.700000</td>\n",
       "      <td>65.8</td>\n",
       "      <td>33.200000</td>\n",
       "      <td>101.40</td>\n",
       "      <td>78.100000</td>\n",
       "      <td>72.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOT2</th>\n",
       "      <td>907.700000</td>\n",
       "      <td>313.500000</td>\n",
       "      <td>415.40</td>\n",
       "      <td>465.50</td>\n",
       "      <td>587.40</td>\n",
       "      <td>403.1</td>\n",
       "      <td>219.600000</td>\n",
       "      <td>256.4</td>\n",
       "      <td>305.500000</td>\n",
       "      <td>361.700000</td>\n",
       "      <td>...</td>\n",
       "      <td>205.000000</td>\n",
       "      <td>299.00</td>\n",
       "      <td>266.3</td>\n",
       "      <td>1301.300000</td>\n",
       "      <td>809.900000</td>\n",
       "      <td>711.1</td>\n",
       "      <td>777.500000</td>\n",
       "      <td>387.70</td>\n",
       "      <td>403.400000</td>\n",
       "      <td>916.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>298.400000</td>\n",
       "      <td>414.150000</td>\n",
       "      <td>866.25</td>\n",
       "      <td>414.35</td>\n",
       "      <td>758.95</td>\n",
       "      <td>509.8</td>\n",
       "      <td>616.750000</td>\n",
       "      <td>516.8</td>\n",
       "      <td>438.100000</td>\n",
       "      <td>596.250000</td>\n",
       "      <td>...</td>\n",
       "      <td>709.850000</td>\n",
       "      <td>408.95</td>\n",
       "      <td>294.8</td>\n",
       "      <td>1057.650000</td>\n",
       "      <td>1041.700000</td>\n",
       "      <td>1492.7</td>\n",
       "      <td>783.600000</td>\n",
       "      <td>793.45</td>\n",
       "      <td>1040.650000</td>\n",
       "      <td>994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>223.833333</td>\n",
       "      <td>59.333333</td>\n",
       "      <td>228.60</td>\n",
       "      <td>138.20</td>\n",
       "      <td>223.80</td>\n",
       "      <td>148.4</td>\n",
       "      <td>336.166667</td>\n",
       "      <td>134.5</td>\n",
       "      <td>160.166667</td>\n",
       "      <td>193.266667</td>\n",
       "      <td>...</td>\n",
       "      <td>132.533333</td>\n",
       "      <td>253.60</td>\n",
       "      <td>270.0</td>\n",
       "      <td>149.333333</td>\n",
       "      <td>364.433333</td>\n",
       "      <td>257.6</td>\n",
       "      <td>231.633333</td>\n",
       "      <td>213.70</td>\n",
       "      <td>136.933333</td>\n",
       "      <td>210.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>835.000000</td>\n",
       "      <td>525.000000</td>\n",
       "      <td>420.60</td>\n",
       "      <td>427.30</td>\n",
       "      <td>655.40</td>\n",
       "      <td>643.8</td>\n",
       "      <td>580.900000</td>\n",
       "      <td>807.2</td>\n",
       "      <td>741.900000</td>\n",
       "      <td>639.900000</td>\n",
       "      <td>...</td>\n",
       "      <td>723.700000</td>\n",
       "      <td>749.90</td>\n",
       "      <td>528.4</td>\n",
       "      <td>275.900000</td>\n",
       "      <td>425.200000</td>\n",
       "      <td>745.9</td>\n",
       "      <td>541.300000</td>\n",
       "      <td>312.50</td>\n",
       "      <td>736.500000</td>\n",
       "      <td>658.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abParts</th>\n",
       "      <td>30.800000</td>\n",
       "      <td>29.900000</td>\n",
       "      <td>57.50</td>\n",
       "      <td>12.10</td>\n",
       "      <td>33.60</td>\n",
       "      <td>19.6</td>\n",
       "      <td>25.600000</td>\n",
       "      <td>40.2</td>\n",
       "      <td>11.900000</td>\n",
       "      <td>10.100000</td>\n",
       "      <td>...</td>\n",
       "      <td>71.200000</td>\n",
       "      <td>28.40</td>\n",
       "      <td>58.0</td>\n",
       "      <td>16.800000</td>\n",
       "      <td>22.800000</td>\n",
       "      <td>31.8</td>\n",
       "      <td>22.600000</td>\n",
       "      <td>53.10</td>\n",
       "      <td>29.500000</td>\n",
       "      <td>133.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>133.900000</td>\n",
       "      <td>153.800000</td>\n",
       "      <td>91.50</td>\n",
       "      <td>128.30</td>\n",
       "      <td>87.60</td>\n",
       "      <td>103.0</td>\n",
       "      <td>297.200000</td>\n",
       "      <td>84.7</td>\n",
       "      <td>118.800000</td>\n",
       "      <td>181.700000</td>\n",
       "      <td>...</td>\n",
       "      <td>190.100000</td>\n",
       "      <td>199.70</td>\n",
       "      <td>177.4</td>\n",
       "      <td>92.700000</td>\n",
       "      <td>190.900000</td>\n",
       "      <td>208.1</td>\n",
       "      <td>135.100000</td>\n",
       "      <td>321.60</td>\n",
       "      <td>201.800000</td>\n",
       "      <td>138.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14828 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM1954726  GSM1954727  GSM1954728  GSM1954729  GSM1954730  \\\n",
       "Gene                                                                    \n",
       " ABCB4     147.500000  695.900000      466.20     1479.40     1155.70   \n",
       " ABCC6P1   957.900000  384.200000      320.80      218.10      445.40   \n",
       " ABCC6P2   957.900000  384.200000      320.80      218.10      445.40   \n",
       " ABCD1P2   171.600000   30.100000       26.20       38.70       34.00   \n",
       " ACOT2     907.700000  313.500000      415.40      465.50      587.40   \n",
       "...               ...         ...         ...         ...         ...   \n",
       "ZYX        298.400000  414.150000      866.25      414.35      758.95   \n",
       "ZZEF1      223.833333   59.333333      228.60      138.20      223.80   \n",
       "ZZZ3       835.000000  525.000000      420.60      427.30      655.40   \n",
       "abParts     30.800000   29.900000       57.50       12.10       33.60   \n",
       "alpha      133.900000  153.800000       91.50      128.30       87.60   \n",
       "\n",
       "           GSM1954731   GSM1954732  GSM1954733  GSM1954734  GSM1954735  ...  \\\n",
       "Gene                                                                    ...   \n",
       " ABCB4          676.0  1514.800000        35.9  202.900000  993.100000  ...   \n",
       " ABCC6P1        238.3   387.700000       495.9  154.200000  241.300000  ...   \n",
       " ABCC6P2        238.3   387.700000       495.9  154.200000  241.300000  ...   \n",
       " ABCD1P2         34.8    59.300000        82.4   30.200000  176.500000  ...   \n",
       " ACOT2          403.1   219.600000       256.4  305.500000  361.700000  ...   \n",
       "...               ...          ...         ...         ...         ...  ...   \n",
       "ZYX             509.8   616.750000       516.8  438.100000  596.250000  ...   \n",
       "ZZEF1           148.4   336.166667       134.5  160.166667  193.266667  ...   \n",
       "ZZZ3            643.8   580.900000       807.2  741.900000  639.900000  ...   \n",
       "abParts          19.6    25.600000        40.2   11.900000   10.100000  ...   \n",
       "alpha           103.0   297.200000        84.7  118.800000  181.700000  ...   \n",
       "\n",
       "           GSM1954747  GSM1954748  GSM1954749   GSM1954750   GSM1954751  \\\n",
       "Gene                                                                      \n",
       " ABCB4     405.800000     1016.00       141.5  6708.400000  1930.100000   \n",
       " ABCC6P1    67.700000      245.70       308.5   118.100000   424.400000   \n",
       " ABCC6P2    67.700000      245.70       308.5   118.100000   424.400000   \n",
       " ABCD1P2    38.700000       78.70        36.6   197.700000    41.700000   \n",
       " ACOT2     205.000000      299.00       266.3  1301.300000   809.900000   \n",
       "...               ...         ...         ...          ...          ...   \n",
       "ZYX        709.850000      408.95       294.8  1057.650000  1041.700000   \n",
       "ZZEF1      132.533333      253.60       270.0   149.333333   364.433333   \n",
       "ZZZ3       723.700000      749.90       528.4   275.900000   425.200000   \n",
       "abParts     71.200000       28.40        58.0    16.800000    22.800000   \n",
       "alpha      190.100000      199.70       177.4    92.700000   190.900000   \n",
       "\n",
       "           GSM1954752   GSM1954753  GSM1954754   GSM1954755  GSM1954756  \n",
       "Gene                                                                     \n",
       " ABCB4         1995.0  2711.300000     3506.00  4596.500000      2504.1  \n",
       " ABCC6P1        235.7   494.700000      563.50   501.000000        82.0  \n",
       " ABCC6P2        235.7   494.700000      563.50   501.000000        82.0  \n",
       " ABCD1P2         65.8    33.200000      101.40    78.100000        72.9  \n",
       " ACOT2          711.1   777.500000      387.70   403.400000       916.1  \n",
       "...               ...          ...         ...          ...         ...  \n",
       "ZYX            1492.7   783.600000      793.45  1040.650000       994.0  \n",
       "ZZEF1           257.6   231.633333      213.70   136.933333       210.3  \n",
       "ZZZ3            745.9   541.300000      312.50   736.500000       658.0  \n",
       "abParts          31.8    22.600000       53.10    29.500000       133.6  \n",
       "alpha           208.1   135.100000      321.60   201.800000       138.1  \n",
       "\n",
       "[14828 rows x 31 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1954726</th>\n",
       "      <th>GSM1954727</th>\n",
       "      <th>GSM1954728</th>\n",
       "      <th>GSM1954729</th>\n",
       "      <th>GSM1954730</th>\n",
       "      <th>GSM1954731</th>\n",
       "      <th>GSM1954732</th>\n",
       "      <th>GSM1954733</th>\n",
       "      <th>GSM1954734</th>\n",
       "      <th>GSM1954735</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1954747</th>\n",
       "      <th>GSM1954748</th>\n",
       "      <th>GSM1954749</th>\n",
       "      <th>GSM1954750</th>\n",
       "      <th>GSM1954751</th>\n",
       "      <th>GSM1954752</th>\n",
       "      <th>GSM1954753</th>\n",
       "      <th>GSM1954754</th>\n",
       "      <th>GSM1954755</th>\n",
       "      <th>GSM1954756</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>435.000000</td>\n",
       "      <td>344.600000</td>\n",
       "      <td>540.80</td>\n",
       "      <td>268.10</td>\n",
       "      <td>435.40</td>\n",
       "      <td>410.7</td>\n",
       "      <td>514.800000</td>\n",
       "      <td>279.2</td>\n",
       "      <td>275.300000</td>\n",
       "      <td>418.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>441.300000</td>\n",
       "      <td>689.40</td>\n",
       "      <td>401.6</td>\n",
       "      <td>381.400000</td>\n",
       "      <td>323.100000</td>\n",
       "      <td>560.4</td>\n",
       "      <td>651.200000</td>\n",
       "      <td>1040.30</td>\n",
       "      <td>655.800000</td>\n",
       "      <td>872.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>4469.000000</td>\n",
       "      <td>7595.800000</td>\n",
       "      <td>5954.90</td>\n",
       "      <td>6257.10</td>\n",
       "      <td>6142.20</td>\n",
       "      <td>4218.6</td>\n",
       "      <td>2760.300000</td>\n",
       "      <td>3700.1</td>\n",
       "      <td>4408.800000</td>\n",
       "      <td>3862.600000</td>\n",
       "      <td>...</td>\n",
       "      <td>2561.200000</td>\n",
       "      <td>2880.80</td>\n",
       "      <td>3540.0</td>\n",
       "      <td>7180.500000</td>\n",
       "      <td>11459.700000</td>\n",
       "      <td>5347.7</td>\n",
       "      <td>4320.200000</td>\n",
       "      <td>5798.30</td>\n",
       "      <td>5584.700000</td>\n",
       "      <td>9061.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>42.600000</td>\n",
       "      <td>61.300000</td>\n",
       "      <td>101.80</td>\n",
       "      <td>140.70</td>\n",
       "      <td>142.60</td>\n",
       "      <td>29.3</td>\n",
       "      <td>279.300000</td>\n",
       "      <td>77.1</td>\n",
       "      <td>55.400000</td>\n",
       "      <td>63.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>76.000000</td>\n",
       "      <td>190.00</td>\n",
       "      <td>208.3</td>\n",
       "      <td>37.400000</td>\n",
       "      <td>97.800000</td>\n",
       "      <td>220.8</td>\n",
       "      <td>88.700000</td>\n",
       "      <td>84.70</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>92.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>352.800000</td>\n",
       "      <td>220.600000</td>\n",
       "      <td>201.60</td>\n",
       "      <td>110.00</td>\n",
       "      <td>181.60</td>\n",
       "      <td>136.6</td>\n",
       "      <td>380.400000</td>\n",
       "      <td>194.5</td>\n",
       "      <td>151.600000</td>\n",
       "      <td>198.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>177.800000</td>\n",
       "      <td>363.60</td>\n",
       "      <td>175.6</td>\n",
       "      <td>154.000000</td>\n",
       "      <td>441.100000</td>\n",
       "      <td>442.7</td>\n",
       "      <td>406.300000</td>\n",
       "      <td>368.10</td>\n",
       "      <td>446.900000</td>\n",
       "      <td>342.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>299.700000</td>\n",
       "      <td>268.300000</td>\n",
       "      <td>481.90</td>\n",
       "      <td>409.70</td>\n",
       "      <td>488.80</td>\n",
       "      <td>330.6</td>\n",
       "      <td>542.700000</td>\n",
       "      <td>715.2</td>\n",
       "      <td>638.300000</td>\n",
       "      <td>1012.200000</td>\n",
       "      <td>...</td>\n",
       "      <td>180.800000</td>\n",
       "      <td>420.70</td>\n",
       "      <td>497.0</td>\n",
       "      <td>311.900000</td>\n",
       "      <td>431.800000</td>\n",
       "      <td>339.8</td>\n",
       "      <td>439.400000</td>\n",
       "      <td>336.40</td>\n",
       "      <td>419.100000</td>\n",
       "      <td>249.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>117.400000</td>\n",
       "      <td>24.400000</td>\n",
       "      <td>15.90</td>\n",
       "      <td>20.10</td>\n",
       "      <td>9.50</td>\n",
       "      <td>52.2</td>\n",
       "      <td>261.400000</td>\n",
       "      <td>16.1</td>\n",
       "      <td>19.100000</td>\n",
       "      <td>113.400000</td>\n",
       "      <td>...</td>\n",
       "      <td>110.400000</td>\n",
       "      <td>30.30</td>\n",
       "      <td>100.1</td>\n",
       "      <td>54.800000</td>\n",
       "      <td>34.400000</td>\n",
       "      <td>100.2</td>\n",
       "      <td>189.500000</td>\n",
       "      <td>208.30</td>\n",
       "      <td>19.200000</td>\n",
       "      <td>70.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>135.200000</td>\n",
       "      <td>336.100000</td>\n",
       "      <td>300.30</td>\n",
       "      <td>233.10</td>\n",
       "      <td>289.70</td>\n",
       "      <td>234.8</td>\n",
       "      <td>108.600000</td>\n",
       "      <td>197.9</td>\n",
       "      <td>464.800000</td>\n",
       "      <td>308.600000</td>\n",
       "      <td>...</td>\n",
       "      <td>198.900000</td>\n",
       "      <td>161.20</td>\n",
       "      <td>291.1</td>\n",
       "      <td>203.400000</td>\n",
       "      <td>204.900000</td>\n",
       "      <td>238.4</td>\n",
       "      <td>63.400000</td>\n",
       "      <td>47.50</td>\n",
       "      <td>274.200000</td>\n",
       "      <td>98.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>298.400000</td>\n",
       "      <td>414.150000</td>\n",
       "      <td>866.25</td>\n",
       "      <td>414.35</td>\n",
       "      <td>758.95</td>\n",
       "      <td>509.8</td>\n",
       "      <td>616.750000</td>\n",
       "      <td>516.8</td>\n",
       "      <td>438.100000</td>\n",
       "      <td>596.250000</td>\n",
       "      <td>...</td>\n",
       "      <td>709.850000</td>\n",
       "      <td>408.95</td>\n",
       "      <td>294.8</td>\n",
       "      <td>1057.650000</td>\n",
       "      <td>1041.700000</td>\n",
       "      <td>1492.7</td>\n",
       "      <td>783.600000</td>\n",
       "      <td>793.45</td>\n",
       "      <td>1040.650000</td>\n",
       "      <td>994.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>223.833333</td>\n",
       "      <td>59.333333</td>\n",
       "      <td>228.60</td>\n",
       "      <td>138.20</td>\n",
       "      <td>223.80</td>\n",
       "      <td>148.4</td>\n",
       "      <td>336.166667</td>\n",
       "      <td>134.5</td>\n",
       "      <td>160.166667</td>\n",
       "      <td>193.266667</td>\n",
       "      <td>...</td>\n",
       "      <td>132.533333</td>\n",
       "      <td>253.60</td>\n",
       "      <td>270.0</td>\n",
       "      <td>149.333333</td>\n",
       "      <td>364.433333</td>\n",
       "      <td>257.6</td>\n",
       "      <td>231.633333</td>\n",
       "      <td>213.70</td>\n",
       "      <td>136.933333</td>\n",
       "      <td>210.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>835.000000</td>\n",
       "      <td>525.000000</td>\n",
       "      <td>420.60</td>\n",
       "      <td>427.30</td>\n",
       "      <td>655.40</td>\n",
       "      <td>643.8</td>\n",
       "      <td>580.900000</td>\n",
       "      <td>807.2</td>\n",
       "      <td>741.900000</td>\n",
       "      <td>639.900000</td>\n",
       "      <td>...</td>\n",
       "      <td>723.700000</td>\n",
       "      <td>749.90</td>\n",
       "      <td>528.4</td>\n",
       "      <td>275.900000</td>\n",
       "      <td>425.200000</td>\n",
       "      <td>745.9</td>\n",
       "      <td>541.300000</td>\n",
       "      <td>312.50</td>\n",
       "      <td>736.500000</td>\n",
       "      <td>658.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11727 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1954726   GSM1954727  GSM1954728  GSM1954729  GSM1954730  \\\n",
       "A1CF     435.000000   344.600000      540.80      268.10      435.40   \n",
       "A2M     4469.000000  7595.800000     5954.90     6257.10     6142.20   \n",
       "A4GALT    42.600000    61.300000      101.80      140.70      142.60   \n",
       "A4GNT    352.800000   220.600000      201.60      110.00      181.60   \n",
       "AAAS     299.700000   268.300000      481.90      409.70      488.80   \n",
       "...             ...          ...         ...         ...         ...   \n",
       "ZXDB     117.400000    24.400000       15.90       20.10        9.50   \n",
       "ZXDC     135.200000   336.100000      300.30      233.10      289.70   \n",
       "ZYX      298.400000   414.150000      866.25      414.35      758.95   \n",
       "ZZEF1    223.833333    59.333333      228.60      138.20      223.80   \n",
       "ZZZ3     835.000000   525.000000      420.60      427.30      655.40   \n",
       "\n",
       "        GSM1954731   GSM1954732  GSM1954733   GSM1954734   GSM1954735  ...  \\\n",
       "A1CF         410.7   514.800000       279.2   275.300000   418.300000  ...   \n",
       "A2M         4218.6  2760.300000      3700.1  4408.800000  3862.600000  ...   \n",
       "A4GALT        29.3   279.300000        77.1    55.400000    63.300000  ...   \n",
       "A4GNT        136.6   380.400000       194.5   151.600000   198.000000  ...   \n",
       "AAAS         330.6   542.700000       715.2   638.300000  1012.200000  ...   \n",
       "...            ...          ...         ...          ...          ...  ...   \n",
       "ZXDB          52.2   261.400000        16.1    19.100000   113.400000  ...   \n",
       "ZXDC         234.8   108.600000       197.9   464.800000   308.600000  ...   \n",
       "ZYX          509.8   616.750000       516.8   438.100000   596.250000  ...   \n",
       "ZZEF1        148.4   336.166667       134.5   160.166667   193.266667  ...   \n",
       "ZZZ3         643.8   580.900000       807.2   741.900000   639.900000  ...   \n",
       "\n",
       "         GSM1954747  GSM1954748  GSM1954749   GSM1954750    GSM1954751  \\\n",
       "A1CF     441.300000      689.40       401.6   381.400000    323.100000   \n",
       "A2M     2561.200000     2880.80      3540.0  7180.500000  11459.700000   \n",
       "A4GALT    76.000000      190.00       208.3    37.400000     97.800000   \n",
       "A4GNT    177.800000      363.60       175.6   154.000000    441.100000   \n",
       "AAAS     180.800000      420.70       497.0   311.900000    431.800000   \n",
       "...             ...         ...         ...          ...           ...   \n",
       "ZXDB     110.400000       30.30       100.1    54.800000     34.400000   \n",
       "ZXDC     198.900000      161.20       291.1   203.400000    204.900000   \n",
       "ZYX      709.850000      408.95       294.8  1057.650000   1041.700000   \n",
       "ZZEF1    132.533333      253.60       270.0   149.333333    364.433333   \n",
       "ZZZ3     723.700000      749.90       528.4   275.900000    425.200000   \n",
       "\n",
       "        GSM1954752   GSM1954753  GSM1954754   GSM1954755  GSM1954756  \n",
       "A1CF         560.4   651.200000     1040.30   655.800000       872.7  \n",
       "A2M         5347.7  4320.200000     5798.30  5584.700000      9061.2  \n",
       "A4GALT       220.8    88.700000       84.70    28.000000        92.7  \n",
       "A4GNT        442.7   406.300000      368.10   446.900000       342.7  \n",
       "AAAS         339.8   439.400000      336.40   419.100000       249.6  \n",
       "...            ...          ...         ...          ...         ...  \n",
       "ZXDB         100.2   189.500000      208.30    19.200000        70.5  \n",
       "ZXDC         238.4    63.400000       47.50   274.200000        98.3  \n",
       "ZYX         1492.7   783.600000      793.45  1040.650000       994.0  \n",
       "ZZEF1        257.6   231.633333      213.70   136.933333       210.3  \n",
       "ZZZ3         745.9   541.300000      312.50   736.500000       658.0  \n",
       "\n",
       "[11727 rows x 31 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:718: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n"
     ]
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZSWIM8</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1954726</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>435.0</td>\n",
       "      <td>4469.0</td>\n",
       "      <td>42.6</td>\n",
       "      <td>352.8</td>\n",
       "      <td>299.7</td>\n",
       "      <td>625.0</td>\n",
       "      <td>39.5</td>\n",
       "      <td>210.4</td>\n",
       "      <td>...</td>\n",
       "      <td>216.9</td>\n",
       "      <td>690.9</td>\n",
       "      <td>452.9</td>\n",
       "      <td>37.6</td>\n",
       "      <td>418.0</td>\n",
       "      <td>117.4</td>\n",
       "      <td>135.2</td>\n",
       "      <td>298.4</td>\n",
       "      <td>223.833333</td>\n",
       "      <td>835.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954727</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>344.6</td>\n",
       "      <td>7595.8</td>\n",
       "      <td>61.3</td>\n",
       "      <td>220.6</td>\n",
       "      <td>268.3</td>\n",
       "      <td>650.2</td>\n",
       "      <td>323.2</td>\n",
       "      <td>297.45</td>\n",
       "      <td>...</td>\n",
       "      <td>309.1</td>\n",
       "      <td>338.35</td>\n",
       "      <td>487.8</td>\n",
       "      <td>256.0</td>\n",
       "      <td>1331.1</td>\n",
       "      <td>24.4</td>\n",
       "      <td>336.1</td>\n",
       "      <td>414.15</td>\n",
       "      <td>59.333333</td>\n",
       "      <td>525.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954728</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>540.8</td>\n",
       "      <td>5954.9</td>\n",
       "      <td>101.8</td>\n",
       "      <td>201.6</td>\n",
       "      <td>481.9</td>\n",
       "      <td>813.1</td>\n",
       "      <td>597.2</td>\n",
       "      <td>367.55</td>\n",
       "      <td>...</td>\n",
       "      <td>215.8</td>\n",
       "      <td>915.85</td>\n",
       "      <td>300.9</td>\n",
       "      <td>200.8</td>\n",
       "      <td>651.5</td>\n",
       "      <td>15.9</td>\n",
       "      <td>300.3</td>\n",
       "      <td>866.25</td>\n",
       "      <td>228.6</td>\n",
       "      <td>420.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954729</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>268.1</td>\n",
       "      <td>6257.1</td>\n",
       "      <td>140.7</td>\n",
       "      <td>110.0</td>\n",
       "      <td>409.7</td>\n",
       "      <td>3292.3</td>\n",
       "      <td>1980.8</td>\n",
       "      <td>266.05</td>\n",
       "      <td>...</td>\n",
       "      <td>34.3</td>\n",
       "      <td>518.05</td>\n",
       "      <td>379.8</td>\n",
       "      <td>320.5</td>\n",
       "      <td>1436.8</td>\n",
       "      <td>20.1</td>\n",
       "      <td>233.1</td>\n",
       "      <td>414.35</td>\n",
       "      <td>138.2</td>\n",
       "      <td>427.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954730</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>435.4</td>\n",
       "      <td>6142.2</td>\n",
       "      <td>142.6</td>\n",
       "      <td>181.6</td>\n",
       "      <td>488.8</td>\n",
       "      <td>870.8</td>\n",
       "      <td>1894.6</td>\n",
       "      <td>267.3</td>\n",
       "      <td>...</td>\n",
       "      <td>222.2</td>\n",
       "      <td>671.9</td>\n",
       "      <td>309.6</td>\n",
       "      <td>146.6</td>\n",
       "      <td>365.9</td>\n",
       "      <td>9.5</td>\n",
       "      <td>289.7</td>\n",
       "      <td>758.95</td>\n",
       "      <td>223.8</td>\n",
       "      <td>655.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954731</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>410.7</td>\n",
       "      <td>4218.6</td>\n",
       "      <td>29.3</td>\n",
       "      <td>136.6</td>\n",
       "      <td>330.6</td>\n",
       "      <td>2399.1</td>\n",
       "      <td>1062.9</td>\n",
       "      <td>545.9</td>\n",
       "      <td>...</td>\n",
       "      <td>220.4</td>\n",
       "      <td>355.0</td>\n",
       "      <td>442.8</td>\n",
       "      <td>288.1</td>\n",
       "      <td>615.4</td>\n",
       "      <td>52.2</td>\n",
       "      <td>234.8</td>\n",
       "      <td>509.8</td>\n",
       "      <td>148.4</td>\n",
       "      <td>643.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954732</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>514.8</td>\n",
       "      <td>2760.3</td>\n",
       "      <td>279.3</td>\n",
       "      <td>380.4</td>\n",
       "      <td>542.7</td>\n",
       "      <td>1012.9</td>\n",
       "      <td>213.4</td>\n",
       "      <td>68.8</td>\n",
       "      <td>...</td>\n",
       "      <td>529.3</td>\n",
       "      <td>878.35</td>\n",
       "      <td>388.2</td>\n",
       "      <td>69.9</td>\n",
       "      <td>840.8</td>\n",
       "      <td>261.4</td>\n",
       "      <td>108.6</td>\n",
       "      <td>616.75</td>\n",
       "      <td>336.166667</td>\n",
       "      <td>580.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954733</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>279.2</td>\n",
       "      <td>3700.1</td>\n",
       "      <td>77.1</td>\n",
       "      <td>194.5</td>\n",
       "      <td>715.2</td>\n",
       "      <td>1772.3</td>\n",
       "      <td>21.3</td>\n",
       "      <td>319.65</td>\n",
       "      <td>...</td>\n",
       "      <td>163.1</td>\n",
       "      <td>326.2</td>\n",
       "      <td>455.2</td>\n",
       "      <td>715.3</td>\n",
       "      <td>3433.5</td>\n",
       "      <td>16.1</td>\n",
       "      <td>197.9</td>\n",
       "      <td>516.8</td>\n",
       "      <td>134.5</td>\n",
       "      <td>807.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954734</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>275.3</td>\n",
       "      <td>4408.8</td>\n",
       "      <td>55.4</td>\n",
       "      <td>151.6</td>\n",
       "      <td>638.3</td>\n",
       "      <td>457.7</td>\n",
       "      <td>32.6</td>\n",
       "      <td>301.3</td>\n",
       "      <td>...</td>\n",
       "      <td>306.0</td>\n",
       "      <td>1865.9</td>\n",
       "      <td>334.9</td>\n",
       "      <td>261.7</td>\n",
       "      <td>2088.8</td>\n",
       "      <td>19.1</td>\n",
       "      <td>464.8</td>\n",
       "      <td>438.1</td>\n",
       "      <td>160.166667</td>\n",
       "      <td>741.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954735</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>418.3</td>\n",
       "      <td>3862.6</td>\n",
       "      <td>63.3</td>\n",
       "      <td>198.0</td>\n",
       "      <td>1012.2</td>\n",
       "      <td>2131.8</td>\n",
       "      <td>29.1</td>\n",
       "      <td>266.85</td>\n",
       "      <td>...</td>\n",
       "      <td>305.5</td>\n",
       "      <td>1125.05</td>\n",
       "      <td>565.3</td>\n",
       "      <td>589.1</td>\n",
       "      <td>2288.8</td>\n",
       "      <td>113.4</td>\n",
       "      <td>308.6</td>\n",
       "      <td>596.25</td>\n",
       "      <td>193.266667</td>\n",
       "      <td>639.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954736</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>215.3</td>\n",
       "      <td>6353.2</td>\n",
       "      <td>43.1</td>\n",
       "      <td>174.1</td>\n",
       "      <td>777.5</td>\n",
       "      <td>1586.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>237.9</td>\n",
       "      <td>...</td>\n",
       "      <td>88.8</td>\n",
       "      <td>1003.35</td>\n",
       "      <td>380.9</td>\n",
       "      <td>249.4</td>\n",
       "      <td>1088.7</td>\n",
       "      <td>29.3</td>\n",
       "      <td>176.2</td>\n",
       "      <td>764.3</td>\n",
       "      <td>172.9</td>\n",
       "      <td>475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954737</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>309.9</td>\n",
       "      <td>8582.9</td>\n",
       "      <td>208.2</td>\n",
       "      <td>87.3</td>\n",
       "      <td>341.8</td>\n",
       "      <td>580.9</td>\n",
       "      <td>25.2</td>\n",
       "      <td>320.35</td>\n",
       "      <td>...</td>\n",
       "      <td>268.6</td>\n",
       "      <td>464.8</td>\n",
       "      <td>555.1</td>\n",
       "      <td>323.2</td>\n",
       "      <td>2588.8</td>\n",
       "      <td>42.7</td>\n",
       "      <td>512.6</td>\n",
       "      <td>379.35</td>\n",
       "      <td>95.466667</td>\n",
       "      <td>652.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954738</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>210.8</td>\n",
       "      <td>3484.6</td>\n",
       "      <td>120.9</td>\n",
       "      <td>216.2</td>\n",
       "      <td>473.2</td>\n",
       "      <td>429.0</td>\n",
       "      <td>137.5</td>\n",
       "      <td>121.9</td>\n",
       "      <td>...</td>\n",
       "      <td>219.1</td>\n",
       "      <td>906.5</td>\n",
       "      <td>332.1</td>\n",
       "      <td>88.7</td>\n",
       "      <td>1449.4</td>\n",
       "      <td>74.3</td>\n",
       "      <td>420.7</td>\n",
       "      <td>564.2</td>\n",
       "      <td>162.233333</td>\n",
       "      <td>871.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954739</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>372.6</td>\n",
       "      <td>5179.7</td>\n",
       "      <td>215.4</td>\n",
       "      <td>107.4</td>\n",
       "      <td>417.2</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>77.2</td>\n",
       "      <td>301.45</td>\n",
       "      <td>...</td>\n",
       "      <td>231.3</td>\n",
       "      <td>1329.25</td>\n",
       "      <td>494.1</td>\n",
       "      <td>313.5</td>\n",
       "      <td>994.5</td>\n",
       "      <td>61.2</td>\n",
       "      <td>167.9</td>\n",
       "      <td>899.05</td>\n",
       "      <td>251.6</td>\n",
       "      <td>314.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954740</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>448.5</td>\n",
       "      <td>2484.0</td>\n",
       "      <td>62.6</td>\n",
       "      <td>204.3</td>\n",
       "      <td>254.2</td>\n",
       "      <td>1068.2</td>\n",
       "      <td>1380.5</td>\n",
       "      <td>175.85</td>\n",
       "      <td>...</td>\n",
       "      <td>343.2</td>\n",
       "      <td>911.7</td>\n",
       "      <td>523.6</td>\n",
       "      <td>227.4</td>\n",
       "      <td>1122.4</td>\n",
       "      <td>90.7</td>\n",
       "      <td>188.7</td>\n",
       "      <td>713.5</td>\n",
       "      <td>183.766667</td>\n",
       "      <td>980.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954741</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>624.1</td>\n",
       "      <td>3308.5</td>\n",
       "      <td>69.9</td>\n",
       "      <td>260.2</td>\n",
       "      <td>354.3</td>\n",
       "      <td>357.7</td>\n",
       "      <td>224.4</td>\n",
       "      <td>262.85</td>\n",
       "      <td>...</td>\n",
       "      <td>371.3</td>\n",
       "      <td>1669.55</td>\n",
       "      <td>383.1</td>\n",
       "      <td>153.8</td>\n",
       "      <td>447.9</td>\n",
       "      <td>8.2</td>\n",
       "      <td>66.2</td>\n",
       "      <td>1186.65</td>\n",
       "      <td>262.6</td>\n",
       "      <td>541.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954742</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>494.0</td>\n",
       "      <td>5352.8</td>\n",
       "      <td>72.7</td>\n",
       "      <td>378.9</td>\n",
       "      <td>553.7</td>\n",
       "      <td>629.8</td>\n",
       "      <td>41.9</td>\n",
       "      <td>231.15</td>\n",
       "      <td>...</td>\n",
       "      <td>484.2</td>\n",
       "      <td>971.05</td>\n",
       "      <td>520.6</td>\n",
       "      <td>21.1</td>\n",
       "      <td>387.6</td>\n",
       "      <td>81.5</td>\n",
       "      <td>136.4</td>\n",
       "      <td>902.2</td>\n",
       "      <td>201.133333</td>\n",
       "      <td>392.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954743</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>573.7</td>\n",
       "      <td>7882.9</td>\n",
       "      <td>322.2</td>\n",
       "      <td>435.8</td>\n",
       "      <td>269.9</td>\n",
       "      <td>444.7</td>\n",
       "      <td>164.9</td>\n",
       "      <td>217.45</td>\n",
       "      <td>...</td>\n",
       "      <td>247.3</td>\n",
       "      <td>828.3</td>\n",
       "      <td>899.9</td>\n",
       "      <td>37.5</td>\n",
       "      <td>2242.0</td>\n",
       "      <td>289.2</td>\n",
       "      <td>278.4</td>\n",
       "      <td>761.0</td>\n",
       "      <td>187.966667</td>\n",
       "      <td>449.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954744</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>5790.4</td>\n",
       "      <td>200.8</td>\n",
       "      <td>292.9</td>\n",
       "      <td>94.2</td>\n",
       "      <td>377.2</td>\n",
       "      <td>108.6</td>\n",
       "      <td>246.2</td>\n",
       "      <td>...</td>\n",
       "      <td>250.0</td>\n",
       "      <td>617.2</td>\n",
       "      <td>311.3</td>\n",
       "      <td>79.5</td>\n",
       "      <td>1678.6</td>\n",
       "      <td>141.7</td>\n",
       "      <td>28.3</td>\n",
       "      <td>529.3</td>\n",
       "      <td>160.233333</td>\n",
       "      <td>564.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954745</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>254.4</td>\n",
       "      <td>7202.9</td>\n",
       "      <td>93.3</td>\n",
       "      <td>277.9</td>\n",
       "      <td>553.0</td>\n",
       "      <td>944.6</td>\n",
       "      <td>83.4</td>\n",
       "      <td>569.85</td>\n",
       "      <td>...</td>\n",
       "      <td>90.1</td>\n",
       "      <td>500.15</td>\n",
       "      <td>486.8</td>\n",
       "      <td>565.4</td>\n",
       "      <td>1930.4</td>\n",
       "      <td>130.0</td>\n",
       "      <td>203.5</td>\n",
       "      <td>434.95</td>\n",
       "      <td>200.1</td>\n",
       "      <td>826.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954746</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>557.5</td>\n",
       "      <td>2552.8</td>\n",
       "      <td>87.7</td>\n",
       "      <td>329.0</td>\n",
       "      <td>310.3</td>\n",
       "      <td>2314.3</td>\n",
       "      <td>48.1</td>\n",
       "      <td>276.65</td>\n",
       "      <td>...</td>\n",
       "      <td>256.0</td>\n",
       "      <td>612.6</td>\n",
       "      <td>768.4</td>\n",
       "      <td>191.9</td>\n",
       "      <td>3713.3</td>\n",
       "      <td>13.5</td>\n",
       "      <td>301.5</td>\n",
       "      <td>465.75</td>\n",
       "      <td>226.866667</td>\n",
       "      <td>382.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954747</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>441.3</td>\n",
       "      <td>2561.2</td>\n",
       "      <td>76.0</td>\n",
       "      <td>177.8</td>\n",
       "      <td>180.8</td>\n",
       "      <td>1765.8</td>\n",
       "      <td>608.0</td>\n",
       "      <td>376.7</td>\n",
       "      <td>...</td>\n",
       "      <td>294.8</td>\n",
       "      <td>561.4</td>\n",
       "      <td>405.1</td>\n",
       "      <td>443.0</td>\n",
       "      <td>1026.1</td>\n",
       "      <td>110.4</td>\n",
       "      <td>198.9</td>\n",
       "      <td>709.85</td>\n",
       "      <td>132.533333</td>\n",
       "      <td>723.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1954748</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>689.4</td>\n",
       "      <td>2880.8</td>\n",
       "      <td>190.0</td>\n",
       "      <td>363.6</td>\n",
       "      <td>420.7</td>\n",
       "      <td>2371.9</td>\n",
       "      <td>49.0</td>\n",
       "      <td>117.95</td>\n",
       "      <td>...</td>\n",
       "      <td>621.9</td>\n",
       "      <td>560.0</td>\n",
       "      <td>333.3</td>\n",
       "      <td>329.2</td>\n",
       "      <td>1126.0</td>\n",
       "      <td>30.3</td>\n",
       "      <td>161.2</td>\n",
       "      <td>408.95</td>\n",
       "      <td>253.6</td>\n",
       "      <td>749.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23 rows Ã— 11729 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer Gender   A1CF     A2M A4GALT  A4GNT    AAAS  \\\n",
       "GSM1954726                   1.0    1.0  435.0  4469.0   42.6  352.8   299.7   \n",
       "GSM1954727                   1.0    1.0  344.6  7595.8   61.3  220.6   268.3   \n",
       "GSM1954728                   1.0    1.0  540.8  5954.9  101.8  201.6   481.9   \n",
       "GSM1954729                   1.0    1.0  268.1  6257.1  140.7  110.0   409.7   \n",
       "GSM1954730                   1.0    1.0  435.4  6142.2  142.6  181.6   488.8   \n",
       "GSM1954731                   1.0    0.0  410.7  4218.6   29.3  136.6   330.6   \n",
       "GSM1954732                   1.0    1.0  514.8  2760.3  279.3  380.4   542.7   \n",
       "GSM1954733                   1.0    0.0  279.2  3700.1   77.1  194.5   715.2   \n",
       "GSM1954734                   1.0    1.0  275.3  4408.8   55.4  151.6   638.3   \n",
       "GSM1954735                   1.0    0.0  418.3  3862.6   63.3  198.0  1012.2   \n",
       "GSM1954736                   1.0    1.0  215.3  6353.2   43.1  174.1   777.5   \n",
       "GSM1954737                   1.0    1.0  309.9  8582.9  208.2   87.3   341.8   \n",
       "GSM1954738                   1.0    0.0  210.8  3484.6  120.9  216.2   473.2   \n",
       "GSM1954739                   1.0    0.0  372.6  5179.7  215.4  107.4   417.2   \n",
       "GSM1954740                   1.0    1.0  448.5  2484.0   62.6  204.3   254.2   \n",
       "GSM1954741                   1.0    0.0  624.1  3308.5   69.9  260.2   354.3   \n",
       "GSM1954742                   1.0    0.0  494.0  5352.8   72.7  378.9   553.7   \n",
       "GSM1954743                   1.0    0.0  573.7  7882.9  322.2  435.8   269.9   \n",
       "GSM1954744                   1.0    1.0  315.0  5790.4  200.8  292.9    94.2   \n",
       "GSM1954745                   1.0    1.0  254.4  7202.9   93.3  277.9   553.0   \n",
       "GSM1954746                   1.0    1.0  557.5  2552.8   87.7  329.0   310.3   \n",
       "GSM1954747                   1.0    1.0  441.3  2561.2   76.0  177.8   180.8   \n",
       "GSM1954748                   1.0    1.0  689.4  2880.8  190.0  363.6   420.7   \n",
       "\n",
       "              AACS   AADAC   AAGAB  ... ZSWIM1   ZSWIM8   ZW10 ZWILCH   ZWINT  \\\n",
       "GSM1954726   625.0    39.5   210.4  ...  216.9    690.9  452.9   37.6   418.0   \n",
       "GSM1954727   650.2   323.2  297.45  ...  309.1   338.35  487.8  256.0  1331.1   \n",
       "GSM1954728   813.1   597.2  367.55  ...  215.8   915.85  300.9  200.8   651.5   \n",
       "GSM1954729  3292.3  1980.8  266.05  ...   34.3   518.05  379.8  320.5  1436.8   \n",
       "GSM1954730   870.8  1894.6   267.3  ...  222.2    671.9  309.6  146.6   365.9   \n",
       "GSM1954731  2399.1  1062.9   545.9  ...  220.4    355.0  442.8  288.1   615.4   \n",
       "GSM1954732  1012.9   213.4    68.8  ...  529.3   878.35  388.2   69.9   840.8   \n",
       "GSM1954733  1772.3    21.3  319.65  ...  163.1    326.2  455.2  715.3  3433.5   \n",
       "GSM1954734   457.7    32.6   301.3  ...  306.0   1865.9  334.9  261.7  2088.8   \n",
       "GSM1954735  2131.8    29.1  266.85  ...  305.5  1125.05  565.3  589.1  2288.8   \n",
       "GSM1954736  1586.0    20.0   237.9  ...   88.8  1003.35  380.9  249.4  1088.7   \n",
       "GSM1954737   580.9    25.2  320.35  ...  268.6    464.8  555.1  323.2  2588.8   \n",
       "GSM1954738   429.0   137.5   121.9  ...  219.1    906.5  332.1   88.7  1449.4   \n",
       "GSM1954739  1515.0    77.2  301.45  ...  231.3  1329.25  494.1  313.5   994.5   \n",
       "GSM1954740  1068.2  1380.5  175.85  ...  343.2    911.7  523.6  227.4  1122.4   \n",
       "GSM1954741   357.7   224.4  262.85  ...  371.3  1669.55  383.1  153.8   447.9   \n",
       "GSM1954742   629.8    41.9  231.15  ...  484.2   971.05  520.6   21.1   387.6   \n",
       "GSM1954743   444.7   164.9  217.45  ...  247.3    828.3  899.9   37.5  2242.0   \n",
       "GSM1954744   377.2   108.6   246.2  ...  250.0    617.2  311.3   79.5  1678.6   \n",
       "GSM1954745   944.6    83.4  569.85  ...   90.1   500.15  486.8  565.4  1930.4   \n",
       "GSM1954746  2314.3    48.1  276.65  ...  256.0    612.6  768.4  191.9  3713.3   \n",
       "GSM1954747  1765.8   608.0   376.7  ...  294.8    561.4  405.1  443.0  1026.1   \n",
       "GSM1954748  2371.9    49.0  117.95  ...  621.9    560.0  333.3  329.2  1126.0   \n",
       "\n",
       "             ZXDB   ZXDC      ZYX       ZZEF1   ZZZ3  \n",
       "GSM1954726  117.4  135.2    298.4  223.833333  835.0  \n",
       "GSM1954727   24.4  336.1   414.15   59.333333  525.0  \n",
       "GSM1954728   15.9  300.3   866.25       228.6  420.6  \n",
       "GSM1954729   20.1  233.1   414.35       138.2  427.3  \n",
       "GSM1954730    9.5  289.7   758.95       223.8  655.4  \n",
       "GSM1954731   52.2  234.8    509.8       148.4  643.8  \n",
       "GSM1954732  261.4  108.6   616.75  336.166667  580.9  \n",
       "GSM1954733   16.1  197.9    516.8       134.5  807.2  \n",
       "GSM1954734   19.1  464.8    438.1  160.166667  741.9  \n",
       "GSM1954735  113.4  308.6   596.25  193.266667  639.9  \n",
       "GSM1954736   29.3  176.2    764.3       172.9  475.0  \n",
       "GSM1954737   42.7  512.6   379.35   95.466667  652.6  \n",
       "GSM1954738   74.3  420.7    564.2  162.233333  871.4  \n",
       "GSM1954739   61.2  167.9   899.05       251.6  314.8  \n",
       "GSM1954740   90.7  188.7    713.5  183.766667  980.7  \n",
       "GSM1954741    8.2   66.2  1186.65       262.6  541.2  \n",
       "GSM1954742   81.5  136.4    902.2  201.133333  392.6  \n",
       "GSM1954743  289.2  278.4    761.0  187.966667  449.8  \n",
       "GSM1954744  141.7   28.3    529.3  160.233333  564.4  \n",
       "GSM1954745  130.0  203.5   434.95       200.1  826.4  \n",
       "GSM1954746   13.5  301.5   465.75  226.866667  382.1  \n",
       "GSM1954747  110.4  198.9   709.85  132.533333  723.7  \n",
       "GSM1954748   30.3  161.2   408.95       253.6  749.9  \n",
       "\n",
       "[23 rows x 11729 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 23 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 23 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is severely biased.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 8 occurrences. This represents 34.78% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"caArray_golub-00327: Sanger cell line Affymetrix gene expression project\"\n",
      "!Series_summary\t\"The microarray gene expression pattern was studied using 798 different cancer cell lines. The cancer cell lines are obtained from different centers. Annotation information were provided in the supplementary file.\"\n",
      "!Series_overall_design\t\"golub-00327\"\n",
      "!Series_overall_design\t\"Assay Type: Gene Expression\"\n",
      "!Series_overall_design\t\"Provider: Affymetrix\"\n",
      "!Series_overall_design\t\"Array Designs: HT_HG-U133A\"\n",
      "!Series_overall_design\t\"Organism: Homo sapiens (ncbitax)\"\n",
      "!Series_overall_design\t\"Tissue Sites: leukemia, Urinary tract, Lung, BiliaryTract, Autonomic Ganglion, Thyroid gland, Stomach, Breast, Pancreas, Head and Neck, Lymphoma, Colorectal, Placenta, Liver, Brain, Bone, pleura, Skin, endometrium, Ovary, cervix, Oesophagus, Connective and Soft Tissue, Muscle, Kidney, Prostate, Adrenal Gland, Eye, Testis, Smooth Muscle Tissue, Vulva, Unknow\"\n",
      "!Series_overall_design\t\"Material Types: cell, synthetic_RNA, whole_organism, total_RNA, BVG\"\n",
      "!Series_overall_design\t\"Disease States: M3 acute myeloid leukemia, hairy cell leukemia, transitional cell carcinoma, Adenocarcinoma, B cell lymphoma unspecified, Acute Lymphoblastic Leukemia, blast phase chronic myeloid leukemia, Carcinoma, M6 acute myeloid leukemia, Neuroblastoma, follicular carcinoma, ductal carcinoma, Burkitt Lymphoma, Squamous Cell Carcinoma, M5 acute myeloid leukemia, Mycosis Fungoides and Sezary Syndrome, Acute T-Cell Lymphoblastic Leukemia, Adult T-Cell Leukemia/Lymphoma, M2 Therapy-Related Myeloid Neoplasm, Choriocarcinoma, Plasma Cell Myeloma, Hepatocellular Carcinoma, anaplastic large cell lymphoma, primitive neuroectodermal tumor-medulloblastoma, M4 acute myeloid leukemia, B Acute Lymphoblastic Leukemia, Acute Leukemia of Ambiguous Lineage, Osteosarcoma, Hodgkin Lymphoma, Mesothelioma, chondrosarcoma, Glioblastoma Multiforme, Malignant Melanoma, carcinosarcoma-malignant mesodermal mixed tumor, bronchioloalveolar adenocarcinoma, chronic lymphocytic leukemia-small lymphocytic lymphoma, micropapillary carcinoma, diffuse large B cell lymphoma, myelodysplastic syndrome, giant cell carcinoma, teratoma, multipotential sarcoma, Small Cell Carcinoma, ASTROCYTOMA, Fibrosarcoma, mucoepidermoid carcinoma, Rhabdomyosarcoma, L1 Acute T-Cell Lymphoblastic Leukemia, Glioma, Anaplastic Astrocytoma, Non-small cell carcinoma, Large Cell Carcinoma, mucinous carcinoma, Acute Myeloid Leukemia, malignant fibrous histiocytoma-pleomorphic sarcoma, clear cell carcinoma, B cell lymphoma unspecified, Anaplastic Carcinoma, Ewings sarcoma-peripheral primitive neuroectodermal tumor, undifferentiated carcinoma, Sarcoma, Embryonal Rhabdomyosarcoma, epithelioid sarcoma, renal cell carcinoma, carcinoid-endocrine tumor, Synovial Sarcoma, lymphoid neoplasm, rhabdoid tumor, Refractory Anemia with Excess Blasts, Liposarcoma, biphasic mesothelioma, adrenal cortical carcinoma, adenosquamous carcinoma, L2 Acute T-Cell Lymphoblastic Leukemia, chronic myeloid leukemia, Micropapillary Serous Carcinoma, desmoplastic, acute leukemia, Retinoblastoma, teratocarcinoma, clear cell renal cell carcinoma, Follicular Lymphoma, Wilms Tumor, M7 acute myeloid leukemia, gliosarcoma, embryonal carcinoma, Leiomyosarcoma, medullary carcinoma, granulosa cell tumor, papillary carcinoma, NS Acute Lymphoblastic Leukemia, papillary transitional cell carcinoma, small cell adenocarcinoma, epithelial dysplasia, hyperplasia, tubular adenocarcinoma, metaplasia, papillary ductal carcinoma, chronic eosinophilic leukemia-hypereosinophilic syndrome, #N/A, malignant trichilemmal cyst, Medullary Breast Carcinoma, L2 Acute Lymphoblastic Leukemia, Osteoblastic Osteosarcoma\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1687570</th>\n",
       "      <th>GSM1687571</th>\n",
       "      <th>GSM1687572</th>\n",
       "      <th>GSM1687573</th>\n",
       "      <th>GSM1687574</th>\n",
       "      <th>GSM1687575</th>\n",
       "      <th>GSM1687576</th>\n",
       "      <th>GSM1687577</th>\n",
       "      <th>GSM1687578</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1688358</th>\n",
       "      <th>GSM1688359</th>\n",
       "      <th>GSM1688360</th>\n",
       "      <th>GSM1688361</th>\n",
       "      <th>GSM1688362</th>\n",
       "      <th>GSM1688363</th>\n",
       "      <th>GSM1688364</th>\n",
       "      <th>GSM1688365</th>\n",
       "      <th>GSM1688366</th>\n",
       "      <th>GSM1688367</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cosmic id: 924101</td>\n",
       "      <td>cosmic id: 906800</td>\n",
       "      <td>cosmic id: 687452</td>\n",
       "      <td>cosmic id: 924100</td>\n",
       "      <td>cosmic id: 910924</td>\n",
       "      <td>cosmic id: 906798</td>\n",
       "      <td>cosmic id: 906797</td>\n",
       "      <td>cosmic id: 906797</td>\n",
       "      <td>cosmic id: 910922</td>\n",
       "      <td>...</td>\n",
       "      <td>cosmic id: 909781</td>\n",
       "      <td>cosmic id: 909782</td>\n",
       "      <td>cosmic id: 909782</td>\n",
       "      <td>cosmic id: 909784</td>\n",
       "      <td>cosmic id: 909785</td>\n",
       "      <td>cosmic id: 909785</td>\n",
       "      <td>cosmic id: 909904</td>\n",
       "      <td>cosmic id: 909905</td>\n",
       "      <td>cosmic id: 687592</td>\n",
       "      <td>cosmic id: 909907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: L2 Acute Lymphoblastic Leukemia</td>\n",
       "      <td>disease state: NS Acute Lymphoblastic Leukemia</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: adenocarcinoma</td>\n",
       "      <td>disease state: adenocarcinoma</td>\n",
       "      <td>disease state: transitional cell carcinoma</td>\n",
       "      <td>disease state: transitional cell carcinoma</td>\n",
       "      <td>disease state: transitional cell carcinoma</td>\n",
       "      <td>disease state: clear cell renal cell carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: renal cell carcinoma</td>\n",
       "      <td>disease state: retinoblastoma</td>\n",
       "      <td>disease state: retinoblastoma</td>\n",
       "      <td>disease state: malignant melanoma</td>\n",
       "      <td>disease state: follicular lymphoma</td>\n",
       "      <td>disease state: follicular lymphoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: glioblastoma multiforme</td>\n",
       "      <td>disease state: glioblastoma multiforme</td>\n",
       "      <td>disease state: ductal carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease location: Hematopoietic and Lymphoid T...</td>\n",
       "      <td>disease location: Hematopoietic and Lymphoid T...</td>\n",
       "      <td>disease location: bladder</td>\n",
       "      <td>disease location: prostate</td>\n",
       "      <td>disease location: stomach</td>\n",
       "      <td>disease location: ureter</td>\n",
       "      <td>disease location: bladder</td>\n",
       "      <td>disease location: bladder</td>\n",
       "      <td>disease location: kidney</td>\n",
       "      <td>...</td>\n",
       "      <td>disease location: kidney</td>\n",
       "      <td>disease location: retina</td>\n",
       "      <td>disease location: retina</td>\n",
       "      <td>disease location: skin</td>\n",
       "      <td>disease location: lymph node</td>\n",
       "      <td>disease location: lymph node</td>\n",
       "      <td>disease location: pancreas</td>\n",
       "      <td>disease location: brain</td>\n",
       "      <td>disease location: temporal lobe</td>\n",
       "      <td>disease location: breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>organism part: Leukemia</td>\n",
       "      <td>organism part: Leukemia</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Prostate</td>\n",
       "      <td>organism part: Stomach</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Kidney</td>\n",
       "      <td>...</td>\n",
       "      <td>organism part: Kidney</td>\n",
       "      <td>organism part: Eye</td>\n",
       "      <td>organism part: Eye</td>\n",
       "      <td>organism part: Skin</td>\n",
       "      <td>organism part: Lymphoma</td>\n",
       "      <td>organism part: Lymphoma</td>\n",
       "      <td>organism part: Pancreas</td>\n",
       "      <td>organism part: Brain</td>\n",
       "      <td>organism part: Brain</td>\n",
       "      <td>organism part: Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample: 736</td>\n",
       "      <td>sample: 494</td>\n",
       "      <td>sample: 7</td>\n",
       "      <td>sample: 746</td>\n",
       "      <td>sample: 439</td>\n",
       "      <td>sample: 168</td>\n",
       "      <td>sample: 152</td>\n",
       "      <td>sample: 37</td>\n",
       "      <td>sample: 450</td>\n",
       "      <td>...</td>\n",
       "      <td>sample: 470</td>\n",
       "      <td>sample: 246</td>\n",
       "      <td>sample: 246</td>\n",
       "      <td>sample: 714</td>\n",
       "      <td>sample: 482</td>\n",
       "      <td>sample: 49</td>\n",
       "      <td>sample: 234</td>\n",
       "      <td>sample: 41</td>\n",
       "      <td>sample: 397</td>\n",
       "      <td>sample: 726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell line code: 749</td>\n",
       "      <td>cell line code: 493</td>\n",
       "      <td>cell line code: 505</td>\n",
       "      <td>cell line code: 760</td>\n",
       "      <td>cell line code: 437</td>\n",
       "      <td>cell line code: 151</td>\n",
       "      <td>cell line code: 134</td>\n",
       "      <td>cell line code: 134</td>\n",
       "      <td>cell line code: 449</td>\n",
       "      <td>...</td>\n",
       "      <td>cell line code: 469</td>\n",
       "      <td>cell line code: 231</td>\n",
       "      <td>cell line code: 231</td>\n",
       "      <td>cell line code: 727</td>\n",
       "      <td>cell line code: 481</td>\n",
       "      <td>cell line code: 481</td>\n",
       "      <td>cell line code: 219</td>\n",
       "      <td>cell line code: 401</td>\n",
       "      <td>cell line code: 390</td>\n",
       "      <td>cell line code: 738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: Unspecified</td>\n",
       "      <td>...</td>\n",
       "      <td>supplier: HSRRB</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: HSRRB</td>\n",
       "      <td>supplier: HSRRB</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>crna plate: 8</td>\n",
       "      <td>crna plate: 6</td>\n",
       "      <td>crna plate: 11</td>\n",
       "      <td>crna plate: 8</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>crna plate: 2</td>\n",
       "      <td>crna plate: 2</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>...</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>crna plate: 3</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 8</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 3</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 4</td>\n",
       "      <td>crna plate: 8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows Ã— 799 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "5  !Sample_characteristics_ch1   \n",
       "6  !Sample_characteristics_ch1   \n",
       "7  !Sample_characteristics_ch1   \n",
       "8  !Sample_characteristics_ch1   \n",
       "\n",
       "                                          GSM1687570  \\\n",
       "0                                  cosmic id: 924101   \n",
       "1     disease state: L2 Acute Lymphoblastic Leukemia   \n",
       "2  disease location: Hematopoietic and Lymphoid T...   \n",
       "3                            organism part: Leukemia   \n",
       "4                                        sample: 736   \n",
       "5                                cell line code: 749   \n",
       "6                                     supplier: DSMZ   \n",
       "7                                      affy_batch: 1   \n",
       "8                                      crna plate: 8   \n",
       "\n",
       "                                          GSM1687571  \\\n",
       "0                                  cosmic id: 906800   \n",
       "1     disease state: NS Acute Lymphoblastic Leukemia   \n",
       "2  disease location: Hematopoietic and Lymphoid T...   \n",
       "3                            organism part: Leukemia   \n",
       "4                                        sample: 494   \n",
       "5                                cell line code: 493   \n",
       "6                                     supplier: DSMZ   \n",
       "7                                      affy_batch: 1   \n",
       "8                                      crna plate: 6   \n",
       "\n",
       "                     GSM1687572                     GSM1687573  \\\n",
       "0             cosmic id: 687452              cosmic id: 924100   \n",
       "1      disease state: carcinoma  disease state: adenocarcinoma   \n",
       "2     disease location: bladder     disease location: prostate   \n",
       "3  organism part: Urinary tract        organism part: Prostate   \n",
       "4                     sample: 7                    sample: 746   \n",
       "5           cell line code: 505            cell line code: 760   \n",
       "6                supplier: ATCC                 supplier: DSMZ   \n",
       "7                 affy_batch: 2                  affy_batch: 1   \n",
       "8                crna plate: 11                  crna plate: 8   \n",
       "\n",
       "                      GSM1687574                                  GSM1687575  \\\n",
       "0              cosmic id: 910924                           cosmic id: 906798   \n",
       "1  disease state: adenocarcinoma  disease state: transitional cell carcinoma   \n",
       "2      disease location: stomach                    disease location: ureter   \n",
       "3         organism part: Stomach                organism part: Urinary tract   \n",
       "4                    sample: 439                                 sample: 168   \n",
       "5            cell line code: 437                         cell line code: 151   \n",
       "6                 supplier: DSMZ                              supplier: DSMZ   \n",
       "7                  affy_batch: 1                               affy_batch: 1   \n",
       "8                  crna plate: 5                               crna plate: 2   \n",
       "\n",
       "                                   GSM1687576  \\\n",
       "0                           cosmic id: 906797   \n",
       "1  disease state: transitional cell carcinoma   \n",
       "2                   disease location: bladder   \n",
       "3                organism part: Urinary tract   \n",
       "4                                 sample: 152   \n",
       "5                         cell line code: 134   \n",
       "6                              supplier: DSMZ   \n",
       "7                               affy_batch: 1   \n",
       "8                               crna plate: 2   \n",
       "\n",
       "                                   GSM1687577  \\\n",
       "0                           cosmic id: 906797   \n",
       "1  disease state: transitional cell carcinoma   \n",
       "2                   disease location: bladder   \n",
       "3                organism part: Urinary tract   \n",
       "4                                  sample: 37   \n",
       "5                         cell line code: 134   \n",
       "6                              supplier: DSMZ   \n",
       "7                               affy_batch: 2   \n",
       "8                              crna plate: 12   \n",
       "\n",
       "                                       GSM1687578  ...  \\\n",
       "0                               cosmic id: 910922  ...   \n",
       "1  disease state: clear cell renal cell carcinoma  ...   \n",
       "2                        disease location: kidney  ...   \n",
       "3                           organism part: Kidney  ...   \n",
       "4                                     sample: 450  ...   \n",
       "5                             cell line code: 449  ...   \n",
       "6                           supplier: Unspecified  ...   \n",
       "7                                   affy_batch: 1  ...   \n",
       "8                                   crna plate: 5  ...   \n",
       "\n",
       "                            GSM1688358                     GSM1688359  \\\n",
       "0                    cosmic id: 909781              cosmic id: 909782   \n",
       "1  disease state: renal cell carcinoma  disease state: retinoblastoma   \n",
       "2             disease location: kidney       disease location: retina   \n",
       "3                organism part: Kidney             organism part: Eye   \n",
       "4                          sample: 470                    sample: 246   \n",
       "5                  cell line code: 469            cell line code: 231   \n",
       "6                      supplier: HSRRB                 supplier: ATCC   \n",
       "7                        affy_batch: 1                  affy_batch: 1   \n",
       "8                        crna plate: 5                  crna plate: 3   \n",
       "\n",
       "                      GSM1688360                         GSM1688361  \\\n",
       "0              cosmic id: 909782                  cosmic id: 909784   \n",
       "1  disease state: retinoblastoma  disease state: malignant melanoma   \n",
       "2       disease location: retina             disease location: skin   \n",
       "3             organism part: Eye                organism part: Skin   \n",
       "4                    sample: 246                        sample: 714   \n",
       "5            cell line code: 231                cell line code: 727   \n",
       "6                 supplier: ATCC                     supplier: ATCC   \n",
       "7                  affy_batch: 2                      affy_batch: 1   \n",
       "8                 crna plate: 12                      crna plate: 8   \n",
       "\n",
       "                           GSM1688362                          GSM1688363  \\\n",
       "0                   cosmic id: 909785                   cosmic id: 909785   \n",
       "1  disease state: follicular lymphoma  disease state: follicular lymphoma   \n",
       "2        disease location: lymph node        disease location: lymph node   \n",
       "3             organism part: Lymphoma             organism part: Lymphoma   \n",
       "4                         sample: 482                          sample: 49   \n",
       "5                 cell line code: 481                 cell line code: 481   \n",
       "6                      supplier: DSMZ                      supplier: DSMZ   \n",
       "7                       affy_batch: 1                       affy_batch: 2   \n",
       "8                       crna plate: 5                      crna plate: 12   \n",
       "\n",
       "                   GSM1688364                              GSM1688365  \\\n",
       "0           cosmic id: 909904                       cosmic id: 909905   \n",
       "1    disease state: carcinoma  disease state: glioblastoma multiforme   \n",
       "2  disease location: pancreas                 disease location: brain   \n",
       "3     organism part: Pancreas                    organism part: Brain   \n",
       "4                 sample: 234                              sample: 41   \n",
       "5         cell line code: 219                     cell line code: 401   \n",
       "6              supplier: DSMZ                         supplier: HSRRB   \n",
       "7               affy_batch: 1                           affy_batch: 2   \n",
       "8               crna plate: 3                          crna plate: 12   \n",
       "\n",
       "                               GSM1688366                       GSM1688367  \n",
       "0                       cosmic id: 687592                cosmic id: 909907  \n",
       "1  disease state: glioblastoma multiforme  disease state: ductal carcinoma  \n",
       "2         disease location: temporal lobe         disease location: breast  \n",
       "3                    organism part: Brain            organism part: Breast  \n",
       "4                             sample: 397                      sample: 726  \n",
       "5                     cell line code: 390              cell line code: 738  \n",
       "6                         supplier: HSRRB                   supplier: ATCC  \n",
       "7                           affy_batch: 1                    affy_batch: 1  \n",
       "8                           crna plate: 4                    crna plate: 8  \n",
       "\n",
       "[9 rows x 799 columns]"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE68950\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'organism part: Leukemia',\n",
       "       'organism part: Urinary tract', 'organism part: Prostate',\n",
       "       'organism part: Stomach', 'organism part: Kidney',\n",
       "       'organism part: Thyroid Gland', 'organism part: Brain',\n",
       "       'organism part: Skin', 'organism part: Muscle',\n",
       "       'organism part: Head and Neck', 'organism part: Ovary',\n",
       "       'organism part: Lung', 'organism part: Autonomic Ganglion',\n",
       "       'organism part: Endometrium', 'organism part: Pancreas',\n",
       "       'organism part: Cervix', 'organism part: Breast',\n",
       "       'organism part: Colorectal', 'organism part: Liver',\n",
       "       'organism part: Vulva', 'organism part: Bone',\n",
       "       'organism part: Oesophagus', 'organism part: BiliaryTract',\n",
       "       'organism part: Connective and Soft Tissue',\n",
       "       'organism part: Lymphoma', 'organism part: Pleura',\n",
       "       'organism part: Testis', 'organism part: Placenta',\n",
       "       'organism part: Adrenal Gland', 'organism part: Unknow',\n",
       "       'organism part: Smooth Muscle Tissue', 'organism part: Eye'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[3]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3\n",
       "organism part: Lung                          139\n",
       "organism part: Leukemia                      116\n",
       "organism part: Brain                          60\n",
       "organism part: Skin                           46\n",
       "organism part: Breast                         44\n",
       "organism part: Autonomic Ganglion             41\n",
       "organism part: Colorectal                     40\n",
       "organism part: Bone                           32\n",
       "organism part: Head and Neck                  25\n",
       "organism part: Oesophagus                     25\n",
       "organism part: Stomach                        25\n",
       "organism part: Ovary                          24\n",
       "organism part: Kidney                         24\n",
       "organism part: Lymphoma                       20\n",
       "organism part: Urinary tract                  20\n",
       "organism part: Pancreas                       17\n",
       "organism part: Cervix                         14\n",
       "organism part: Thyroid Gland                  13\n",
       "organism part: Liver                          11\n",
       "organism part: Endometrium                    11\n",
       "organism part: Connective and Soft Tissue     10\n",
       "organism part: Muscle                          8\n",
       "organism part: BiliaryTract                    6\n",
       "organism part: Prostate                        6\n",
       "organism part: Pleura                          6\n",
       "organism part: Testis                          4\n",
       "organism part: Vulva                           3\n",
       "organism part: Adrenal Gland                   2\n",
       "organism part: Eye                             2\n",
       "organism part: Placenta                        2\n",
       "organism part: Unknow                          1\n",
       "organism part: Smooth Muscle Tissue            1\n",
       "!Sample_characteristics_ch1                    1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_counts = tumor_stage_row.value_counts()\n",
    "tumor_stage_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 3\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    if tissue_type == 'organism part: Adrenal Gland':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1687570</th>\n",
       "      <th>GSM1687571</th>\n",
       "      <th>GSM1687572</th>\n",
       "      <th>GSM1687573</th>\n",
       "      <th>GSM1687574</th>\n",
       "      <th>GSM1687575</th>\n",
       "      <th>GSM1687576</th>\n",
       "      <th>GSM1687577</th>\n",
       "      <th>GSM1687578</th>\n",
       "      <th>GSM1687579</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1688358</th>\n",
       "      <th>GSM1688359</th>\n",
       "      <th>GSM1688360</th>\n",
       "      <th>GSM1688361</th>\n",
       "      <th>GSM1688362</th>\n",
       "      <th>GSM1688363</th>\n",
       "      <th>GSM1688364</th>\n",
       "      <th>GSM1688365</th>\n",
       "      <th>GSM1688366</th>\n",
       "      <th>GSM1688367</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1687570  GSM1687571  GSM1687572  GSM1687573  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM1687574  GSM1687575  GSM1687576  GSM1687577  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM1687578  GSM1687579  ...  GSM1688358  GSM1688359  \\\n",
       "Adrenocortical Cancer           0           0  ...           0           0   \n",
       "\n",
       "                       GSM1688360  GSM1688361  GSM1688362  GSM1688363  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM1688364  GSM1688365  GSM1688366  GSM1688367  \n",
       "Adrenocortical Cancer           0           0           0           0  \n",
       "\n",
       "[1 rows x 798 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1687570</th>\n",
       "      <th>GSM1687571</th>\n",
       "      <th>GSM1687572</th>\n",
       "      <th>GSM1687573</th>\n",
       "      <th>GSM1687574</th>\n",
       "      <th>GSM1687575</th>\n",
       "      <th>GSM1687576</th>\n",
       "      <th>GSM1687577</th>\n",
       "      <th>GSM1687578</th>\n",
       "      <th>GSM1687579</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1688358</th>\n",
       "      <th>GSM1688359</th>\n",
       "      <th>GSM1688360</th>\n",
       "      <th>GSM1688361</th>\n",
       "      <th>GSM1688362</th>\n",
       "      <th>GSM1688363</th>\n",
       "      <th>GSM1688364</th>\n",
       "      <th>GSM1688365</th>\n",
       "      <th>GSM1688366</th>\n",
       "      <th>GSM1688367</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>282.4</td>\n",
       "      <td>109.1</td>\n",
       "      <td>332.4</td>\n",
       "      <td>519.1</td>\n",
       "      <td>1874.4</td>\n",
       "      <td>358.8</td>\n",
       "      <td>3092.0</td>\n",
       "      <td>858.2</td>\n",
       "      <td>102.6</td>\n",
       "      <td>60.8</td>\n",
       "      <td>...</td>\n",
       "      <td>287.3</td>\n",
       "      <td>433.0</td>\n",
       "      <td>124.4</td>\n",
       "      <td>777.6</td>\n",
       "      <td>55.8</td>\n",
       "      <td>13.4</td>\n",
       "      <td>1777.6</td>\n",
       "      <td>125.8</td>\n",
       "      <td>226.5</td>\n",
       "      <td>1193.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>234.6</td>\n",
       "      <td>225.3</td>\n",
       "      <td>203.9</td>\n",
       "      <td>155.9</td>\n",
       "      <td>135.9</td>\n",
       "      <td>126.3</td>\n",
       "      <td>501.1</td>\n",
       "      <td>128.3</td>\n",
       "      <td>105.1</td>\n",
       "      <td>126.6</td>\n",
       "      <td>...</td>\n",
       "      <td>165.5</td>\n",
       "      <td>340.3</td>\n",
       "      <td>137.3</td>\n",
       "      <td>240.2</td>\n",
       "      <td>159.2</td>\n",
       "      <td>44.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>42.9</td>\n",
       "      <td>129.4</td>\n",
       "      <td>155.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>12.6</td>\n",
       "      <td>85.4</td>\n",
       "      <td>1.9</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>24.1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.9</td>\n",
       "      <td>4.6</td>\n",
       "      <td>18.3</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4</td>\n",
       "      <td>2624.0</td>\n",
       "      <td>580.9</td>\n",
       "      <td>28.9</td>\n",
       "      <td>6.9</td>\n",
       "      <td>1.3</td>\n",
       "      <td>9.3</td>\n",
       "      <td>0.9</td>\n",
       "      <td>2.2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>71.9</td>\n",
       "      <td>102.1</td>\n",
       "      <td>57.6</td>\n",
       "      <td>138.5</td>\n",
       "      <td>443.9</td>\n",
       "      <td>112.5</td>\n",
       "      <td>75.7</td>\n",
       "      <td>18.1</td>\n",
       "      <td>2271.1</td>\n",
       "      <td>403.9</td>\n",
       "      <td>...</td>\n",
       "      <td>1281.2</td>\n",
       "      <td>55.6</td>\n",
       "      <td>18.1</td>\n",
       "      <td>78.8</td>\n",
       "      <td>70.6</td>\n",
       "      <td>16.5</td>\n",
       "      <td>69.4</td>\n",
       "      <td>35.7</td>\n",
       "      <td>40.7</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>0.9</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.7</td>\n",
       "      <td>4.1</td>\n",
       "      <td>5.4</td>\n",
       "      <td>3.1</td>\n",
       "      <td>2.4</td>\n",
       "      <td>0.9</td>\n",
       "      <td>1.7</td>\n",
       "      <td>4.7</td>\n",
       "      <td>...</td>\n",
       "      <td>1.7</td>\n",
       "      <td>351.1</td>\n",
       "      <td>90.7</td>\n",
       "      <td>1.9</td>\n",
       "      <td>4.4</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.4</td>\n",
       "      <td>6.4</td>\n",
       "      <td>1.6</td>\n",
       "      <td>3.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM1687570  GSM1687571  GSM1687572  GSM1687573  GSM1687574  \\\n",
       "ID                                                                      \n",
       "1007_s_at       282.4       109.1       332.4       519.1      1874.4   \n",
       "1053_at         234.6       225.3       203.9       155.9       135.9   \n",
       "117_at           12.6        85.4         1.9         6.0         6.0   \n",
       "121_at           71.9       102.1        57.6       138.5       443.9   \n",
       "1255_g_at         0.9         2.8         0.7         4.1         5.4   \n",
       "\n",
       "           GSM1687575  GSM1687576  GSM1687577  GSM1687578  GSM1687579  ...  \\\n",
       "ID                                                                     ...   \n",
       "1007_s_at       358.8      3092.0       858.2       102.6        60.8  ...   \n",
       "1053_at         126.3       501.1       128.3       105.1       126.6  ...   \n",
       "117_at           24.1         2.0         2.9         4.6        18.3  ...   \n",
       "121_at          112.5        75.7        18.1      2271.1       403.9  ...   \n",
       "1255_g_at         3.1         2.4         0.9         1.7         4.7  ...   \n",
       "\n",
       "           GSM1688358  GSM1688359  GSM1688360  GSM1688361  GSM1688362  \\\n",
       "ID                                                                      \n",
       "1007_s_at       287.3       433.0       124.4       777.6        55.8   \n",
       "1053_at         165.5       340.3       137.3       240.2       159.2   \n",
       "117_at            2.4      2624.0       580.9        28.9         6.9   \n",
       "121_at         1281.2        55.6        18.1        78.8        70.6   \n",
       "1255_g_at         1.7       351.1        90.7         1.9         4.4   \n",
       "\n",
       "           GSM1688363  GSM1688364  GSM1688365  GSM1688366  GSM1688367  \n",
       "ID                                                                     \n",
       "1007_s_at        13.4      1777.6       125.8       226.5      1193.4  \n",
       "1053_at          44.0       390.0        42.9       129.4       155.5  \n",
       "117_at            1.3         9.3         0.9         2.2         2.0  \n",
       "121_at           16.5        69.4        35.7        40.7        67.0  \n",
       "1255_g_at         0.5         1.4         6.4         1.6         3.4  \n",
       "\n",
       "[5 rows x 798 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Mar 8, 2007', 'Mar 8, 2007', 'Mar 8, 2007', 'Mar 8, 2007', 'Mar 8, 2007'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': [nan, nan, nan, nan, nan], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor family, member 1', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box gene 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001954 /// NM_013993 /// NM_013994', 'NM_002914 /// NM_181471', 'NM_002155 /// XM_001134322', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409'], 'Gene Ontology Biological Process': ['0006468 // protein amino acid phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation', '0006260 // DNA replication // inferred from electronic annotation', '0006457 // protein folding // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0006986 // response to unfolded protein // inferred from electronic annotation', '0001656 // metanephros development // inferred from electronic annotation /// 0006183 // GTP biosynthesis // inferred from electronic annotation /// 0006228 // UTP biosynthesis // inferred from electronic annotation /// 0006241 // CTP biosynthesis // inferred from electronic annotation /// 0006350 // transcription // inferred from electronic annotation /// 0009887 // organ morphogenesis // inferred from electronic annotation /// 0030154 // cell differentiation // inferred from electronic annotation /// 0045893 // positive regulation of transcription, DNA-dependent // inferred from sequence or structural similarity /// 0006355 // regulation of transcription, DNA-dependent // inferred from electronic annotation /// 0007275 // development // inferred from electronic annotation /// 0009653 // morphogenesis // traceable author statement', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // traceable author statement /// 0050896 // response to stimulus // inferred from electronic annotation /// 0007601 // visual perception // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005615 // extracellular space // inferred from electronic annotation /// 0005887 // integral to plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral to membrane // inferred from electronic annotation', '0005634 // nucleus // inferred from electronic annotation /// 0005663 // DNA replication factor C complex // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from electronic annotation', nan, '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005667 // transcription factor complex // inferred from electronic annotation', nan], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004674 // protein serine/threonine kinase activity // inferred from electronic annotation /// 0004713 // protein-tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0004872 // receptor activity // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // traceable author statement /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation', '0003700 // transcription factor activity // traceable author statement /// 0004550 // nucleoside diphosphate kinase activity // inferred from electronic annotation /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from sequence or structural similarity /// 0005524 // ATP binding // inferred from electronic annotation /// 0016563 // transcriptional activator activity // inferred from sequence or structural similarity /// 0003677 // DNA binding // inferred from electronic annotation', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // traceable author statement']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1598_g_at', '160020_at', '1729_at', '1773_at', '177_at', '179_at', '1861_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Mar 8, 2007', 'Mar 8, 2007', 'Mar 8, 2007', 'Mar 8, 2007', 'Mar 8, 2007'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': [nan, nan, nan, nan, nan], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor family, member 1', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box gene 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001954 /// NM_013993 /// NM_013994', 'NM_002914 /// NM_181471', 'NM_002155 /// XM_001134322', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409'], 'Gene Ontology Biological Process': ['0006468 // protein amino acid phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation', '0006260 // DNA replication // inferred from electronic annotation', '0006457 // protein folding // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0006986 // response to unfolded protein // inferred from electronic annotation', '0001656 // metanephros development // inferred from electronic annotation /// 0006183 // GTP biosynthesis // inferred from electronic annotation /// 0006228 // UTP biosynthesis // inferred from electronic annotation /// 0006241 // CTP biosynthesis // inferred from electronic annotation /// 0006350 // transcription // inferred from electronic annotation /// 0009887 // organ morphogenesis // inferred from electronic annotation /// 0030154 // cell differentiation // inferred from electronic annotation /// 0045893 // positive regulation of transcription, DNA-dependent // inferred from sequence or structural similarity /// 0006355 // regulation of transcription, DNA-dependent // inferred from electronic annotation /// 0007275 // development // inferred from electronic annotation /// 0009653 // morphogenesis // traceable author statement', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // traceable author statement /// 0050896 // response to stimulus // inferred from electronic annotation /// 0007601 // visual perception // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005615 // extracellular space // inferred from electronic annotation /// 0005887 // integral to plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral to membrane // inferred from electronic annotation', '0005634 // nucleus // inferred from electronic annotation /// 0005663 // DNA replication factor C complex // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from electronic annotation', nan, '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005667 // transcription factor complex // inferred from electronic annotation', nan], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004674 // protein serine/threonine kinase activity // inferred from electronic annotation /// 0004713 // protein-tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0004872 // receptor activity // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // traceable author statement /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation', '0003700 // transcription factor activity // traceable author statement /// 0004550 // nucleoside diphosphate kinase activity // inferred from electronic annotation /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from sequence or structural similarity /// 0005524 // ATP binding // inferred from electronic annotation /// 0016563 // transcriptional activator activity // inferred from sequence or structural similarity /// 0003677 // DNA binding // inferred from electronic annotation', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // traceable author statement']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mar 8, 2007</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor family, member 1</td>\n",
       "      <td>DDR1</td>\n",
       "      <td>780</td>\n",
       "      <td>NM_001954 /// NM_013993 /// NM_013994</td>\n",
       "      <td>0006468 // protein amino acid phosphorylation ...</td>\n",
       "      <td>0005615 // extracellular space // inferred fro...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mar 8, 2007</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_002914 /// NM_181471</td>\n",
       "      <td>0006260 // DNA replication // inferred from el...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mar 8, 2007</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155 /// XM_001134322</td>\n",
       "      <td>0006457 // protein folding // inferred from el...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mar 8, 2007</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box gene 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001656 // metanephros development // inferred...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0003700 // transcription factor activity // tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Mar 8, 2007</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17800116</th>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>283.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17800117</th>\n",
       "      <td>AFFX-BioDn-5_at</td>\n",
       "      <td>3035.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17800118</th>\n",
       "      <td>AFFX-BioDn-3_at</td>\n",
       "      <td>5720.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17800119</th>\n",
       "      <td>AFFX-CreX-5_at</td>\n",
       "      <td>14758.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17800120</th>\n",
       "      <td>AFFX-CreX-3_at</td>\n",
       "      <td>17663.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17800121 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ID   GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0               1007_s_at   U48705     NaN            Homo sapiens   \n",
       "1                 1053_at   M87338     NaN            Homo sapiens   \n",
       "2                  117_at   X51757     NaN            Homo sapiens   \n",
       "3                  121_at   X69699     NaN            Homo sapiens   \n",
       "4               1255_g_at   L36861     NaN            Homo sapiens   \n",
       "...                   ...      ...     ...                     ...   \n",
       "17800116   AFFX-BioC-3_at    283.9     NaN                     NaN   \n",
       "17800117  AFFX-BioDn-5_at   3035.0     NaN                     NaN   \n",
       "17800118  AFFX-BioDn-3_at   5720.6     NaN                     NaN   \n",
       "17800119   AFFX-CreX-5_at  14758.7     NaN                     NaN   \n",
       "17800120   AFFX-CreX-3_at  17663.6     NaN                     NaN   \n",
       "\n",
       "         Annotation Date      Sequence Type  Sequence Source  \\\n",
       "0            Mar 8, 2007  Exemplar sequence              NaN   \n",
       "1            Mar 8, 2007  Exemplar sequence              NaN   \n",
       "2            Mar 8, 2007  Exemplar sequence              NaN   \n",
       "3            Mar 8, 2007  Exemplar sequence              NaN   \n",
       "4            Mar 8, 2007  Exemplar sequence              NaN   \n",
       "...                  ...                ...              ...   \n",
       "17800116             NaN                NaN              NaN   \n",
       "17800117             NaN                NaN              NaN   \n",
       "17800118             NaN                NaN              NaN   \n",
       "17800119             NaN                NaN              NaN   \n",
       "17800120             NaN                NaN              NaN   \n",
       "\n",
       "                                         Target Description  \\\n",
       "0         U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1         M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2         X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3         X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4         L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                     ...   \n",
       "17800116                                                NaN   \n",
       "17800117                                                NaN   \n",
       "17800118                                                NaN   \n",
       "17800119                                                NaN   \n",
       "17800120                                                NaN   \n",
       "\n",
       "         Representative Public ID  \\\n",
       "0                          U48705   \n",
       "1                          M87338   \n",
       "2                          X51757   \n",
       "3                          X69699   \n",
       "4                          L36861   \n",
       "...                           ...   \n",
       "17800116                      NaN   \n",
       "17800117                      NaN   \n",
       "17800118                      NaN   \n",
       "17800119                      NaN   \n",
       "17800120                      NaN   \n",
       "\n",
       "                                           Gene Title Gene Symbol  \\\n",
       "0          discoidin domain receptor family, member 1        DDR1   \n",
       "1         replication factor C (activator 1) 2, 40kDa        RFC2   \n",
       "2                heat shock 70kDa protein 6 (HSP70B')       HSPA6   \n",
       "3                                   paired box gene 8        PAX8   \n",
       "4             guanylate cyclase activator 1A (retina)      GUCA1A   \n",
       "...                                               ...         ...   \n",
       "17800116                                          NaN         NaN   \n",
       "17800117                                          NaN         NaN   \n",
       "17800118                                          NaN         NaN   \n",
       "17800119                                          NaN         NaN   \n",
       "17800120                                          NaN         NaN   \n",
       "\n",
       "         ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0                   780              NM_001954 /// NM_013993 /// NM_013994   \n",
       "1                  5982                            NM_002914 /// NM_181471   \n",
       "2                  3310                         NM_002155 /// XM_001134322   \n",
       "3                  7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                  2978                                          NM_000409   \n",
       "...                 ...                                                ...   \n",
       "17800116            NaN                                                NaN   \n",
       "17800117            NaN                                                NaN   \n",
       "17800118            NaN                                                NaN   \n",
       "17800119            NaN                                                NaN   \n",
       "17800120            NaN                                                NaN   \n",
       "\n",
       "                           Gene Ontology Biological Process  \\\n",
       "0         0006468 // protein amino acid phosphorylation ...   \n",
       "1         0006260 // DNA replication // inferred from el...   \n",
       "2         0006457 // protein folding // inferred from el...   \n",
       "3         0001656 // metanephros development // inferred...   \n",
       "4         0007165 // signal transduction // non-traceabl...   \n",
       "...                                                     ...   \n",
       "17800116                                                NaN   \n",
       "17800117                                                NaN   \n",
       "17800118                                                NaN   \n",
       "17800119                                                NaN   \n",
       "17800120                                                NaN   \n",
       "\n",
       "                           Gene Ontology Cellular Component  \\\n",
       "0         0005615 // extracellular space // inferred fro...   \n",
       "1         0005634 // nucleus // inferred from electronic...   \n",
       "2                                                       NaN   \n",
       "3         0005634 // nucleus // inferred from electronic...   \n",
       "4                                                       NaN   \n",
       "...                                                     ...   \n",
       "17800116                                                NaN   \n",
       "17800117                                                NaN   \n",
       "17800118                                                NaN   \n",
       "17800119                                                NaN   \n",
       "17800120                                                NaN   \n",
       "\n",
       "                           Gene Ontology Molecular Function  \n",
       "0         0000166 // nucleotide binding // inferred from...  \n",
       "1         0000166 // nucleotide binding // inferred from...  \n",
       "2         0000166 // nucleotide binding // inferred from...  \n",
       "3         0003700 // transcription factor activity // tr...  \n",
       "4         0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                     ...  \n",
       "17800116                                                NaN  \n",
       "17800117                                                NaN  \n",
       "17800118                                                NaN  \n",
       "17800119                                                NaN  \n",
       "17800120                                                NaN  \n",
       "\n",
       "[17800121 rows x 16 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1687570</th>\n",
       "      <th>GSM1687571</th>\n",
       "      <th>GSM1687572</th>\n",
       "      <th>GSM1687573</th>\n",
       "      <th>GSM1687574</th>\n",
       "      <th>GSM1687575</th>\n",
       "      <th>GSM1687576</th>\n",
       "      <th>GSM1687577</th>\n",
       "      <th>GSM1687578</th>\n",
       "      <th>GSM1687579</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1688358</th>\n",
       "      <th>GSM1688359</th>\n",
       "      <th>GSM1688360</th>\n",
       "      <th>GSM1688361</th>\n",
       "      <th>GSM1688362</th>\n",
       "      <th>GSM1688363</th>\n",
       "      <th>GSM1688364</th>\n",
       "      <th>GSM1688365</th>\n",
       "      <th>GSM1688366</th>\n",
       "      <th>GSM1688367</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>259.400000</td>\n",
       "      <td>180.500000</td>\n",
       "      <td>4.200000</td>\n",
       "      <td>26.000000</td>\n",
       "      <td>36.500000</td>\n",
       "      <td>50.600000</td>\n",
       "      <td>70.700000</td>\n",
       "      <td>17.90</td>\n",
       "      <td>166.100000</td>\n",
       "      <td>6.800000</td>\n",
       "      <td>...</td>\n",
       "      <td>13.300000</td>\n",
       "      <td>21.400000</td>\n",
       "      <td>14.500000</td>\n",
       "      <td>555.600000</td>\n",
       "      <td>13.600000</td>\n",
       "      <td>9.400000</td>\n",
       "      <td>26.300000</td>\n",
       "      <td>4.200000</td>\n",
       "      <td>23.300000</td>\n",
       "      <td>13.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOT1</th>\n",
       "      <td>54.300000</td>\n",
       "      <td>269.600000</td>\n",
       "      <td>35.900000</td>\n",
       "      <td>63.400000</td>\n",
       "      <td>57.500000</td>\n",
       "      <td>323.600000</td>\n",
       "      <td>53.000000</td>\n",
       "      <td>14.40</td>\n",
       "      <td>35.100000</td>\n",
       "      <td>43.200000</td>\n",
       "      <td>...</td>\n",
       "      <td>39.600000</td>\n",
       "      <td>124.400000</td>\n",
       "      <td>31.500000</td>\n",
       "      <td>201.800000</td>\n",
       "      <td>27.800000</td>\n",
       "      <td>9.500000</td>\n",
       "      <td>227.600000</td>\n",
       "      <td>21.600000</td>\n",
       "      <td>33.200000</td>\n",
       "      <td>441.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM2</th>\n",
       "      <td>8.300000</td>\n",
       "      <td>4.900000</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>5.400000</td>\n",
       "      <td>5.700000</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>13.800000</td>\n",
       "      <td>8.10</td>\n",
       "      <td>102.700000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>8.700000</td>\n",
       "      <td>1.900000</td>\n",
       "      <td>2.800000</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>1.800000</td>\n",
       "      <td>14.800000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM3</th>\n",
       "      <td>13418.200000</td>\n",
       "      <td>17818.800000</td>\n",
       "      <td>4635.400000</td>\n",
       "      <td>13449.900000</td>\n",
       "      <td>12204.700000</td>\n",
       "      <td>16411.300000</td>\n",
       "      <td>16341.800000</td>\n",
       "      <td>4182.80</td>\n",
       "      <td>9947.700000</td>\n",
       "      <td>3564.800000</td>\n",
       "      <td>...</td>\n",
       "      <td>9886.500000</td>\n",
       "      <td>10840.700000</td>\n",
       "      <td>3032.000000</td>\n",
       "      <td>10719.200000</td>\n",
       "      <td>8010.500000</td>\n",
       "      <td>3099.900000</td>\n",
       "      <td>13392.000000</td>\n",
       "      <td>3407.400000</td>\n",
       "      <td>8757.700000</td>\n",
       "      <td>11117.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM21P</th>\n",
       "      <td>3.700000</td>\n",
       "      <td>54.000000</td>\n",
       "      <td>4.900000</td>\n",
       "      <td>5.900000</td>\n",
       "      <td>3.700000</td>\n",
       "      <td>36.900000</td>\n",
       "      <td>31.100000</td>\n",
       "      <td>14.70</td>\n",
       "      <td>32.700000</td>\n",
       "      <td>15.800000</td>\n",
       "      <td>...</td>\n",
       "      <td>59.900000</td>\n",
       "      <td>15.500000</td>\n",
       "      <td>4.100000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>3.200000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>17.600000</td>\n",
       "      <td>11.100000</td>\n",
       "      <td>19.700000</td>\n",
       "      <td>25.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11BL</th>\n",
       "      <td>16.200000</td>\n",
       "      <td>40.233333</td>\n",
       "      <td>21.966667</td>\n",
       "      <td>42.900000</td>\n",
       "      <td>30.533333</td>\n",
       "      <td>44.800000</td>\n",
       "      <td>90.466667</td>\n",
       "      <td>26.20</td>\n",
       "      <td>34.166667</td>\n",
       "      <td>22.400000</td>\n",
       "      <td>...</td>\n",
       "      <td>44.466667</td>\n",
       "      <td>31.666667</td>\n",
       "      <td>11.966667</td>\n",
       "      <td>19.133333</td>\n",
       "      <td>21.366667</td>\n",
       "      <td>11.933333</td>\n",
       "      <td>67.400000</td>\n",
       "      <td>20.833333</td>\n",
       "      <td>27.266667</td>\n",
       "      <td>19.433333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>54.400000</td>\n",
       "      <td>65.550000</td>\n",
       "      <td>123.800000</td>\n",
       "      <td>112.850000</td>\n",
       "      <td>174.700000</td>\n",
       "      <td>546.100000</td>\n",
       "      <td>433.400000</td>\n",
       "      <td>109.05</td>\n",
       "      <td>678.500000</td>\n",
       "      <td>80.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>488.650000</td>\n",
       "      <td>48.700000</td>\n",
       "      <td>8.500000</td>\n",
       "      <td>125.550000</td>\n",
       "      <td>259.250000</td>\n",
       "      <td>58.500000</td>\n",
       "      <td>318.050000</td>\n",
       "      <td>384.800000</td>\n",
       "      <td>198.850000</td>\n",
       "      <td>66.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>55.566667</td>\n",
       "      <td>98.433333</td>\n",
       "      <td>22.566667</td>\n",
       "      <td>44.566667</td>\n",
       "      <td>61.533333</td>\n",
       "      <td>137.966667</td>\n",
       "      <td>150.300000</td>\n",
       "      <td>27.20</td>\n",
       "      <td>47.433333</td>\n",
       "      <td>48.366667</td>\n",
       "      <td>...</td>\n",
       "      <td>50.700000</td>\n",
       "      <td>79.233333</td>\n",
       "      <td>12.933333</td>\n",
       "      <td>74.133333</td>\n",
       "      <td>36.433333</td>\n",
       "      <td>14.100000</td>\n",
       "      <td>51.133333</td>\n",
       "      <td>16.133333</td>\n",
       "      <td>41.200000</td>\n",
       "      <td>37.966667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>231.000000</td>\n",
       "      <td>420.700000</td>\n",
       "      <td>415.200000</td>\n",
       "      <td>190.400000</td>\n",
       "      <td>415.400000</td>\n",
       "      <td>589.200000</td>\n",
       "      <td>1376.500000</td>\n",
       "      <td>260.10</td>\n",
       "      <td>251.500000</td>\n",
       "      <td>331.100000</td>\n",
       "      <td>...</td>\n",
       "      <td>257.500000</td>\n",
       "      <td>194.500000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>236.600000</td>\n",
       "      <td>120.500000</td>\n",
       "      <td>43.600000</td>\n",
       "      <td>465.400000</td>\n",
       "      <td>139.400000</td>\n",
       "      <td>180.400000</td>\n",
       "      <td>266.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ222E13.2</th>\n",
       "      <td>39.600000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>33.300000</td>\n",
       "      <td>2.400000</td>\n",
       "      <td>32.500000</td>\n",
       "      <td>94.500000</td>\n",
       "      <td>86.400000</td>\n",
       "      <td>30.30</td>\n",
       "      <td>5.700000</td>\n",
       "      <td>4.600000</td>\n",
       "      <td>...</td>\n",
       "      <td>35.500000</td>\n",
       "      <td>41.600000</td>\n",
       "      <td>14.300000</td>\n",
       "      <td>48.500000</td>\n",
       "      <td>17.700000</td>\n",
       "      <td>1.800000</td>\n",
       "      <td>59.500000</td>\n",
       "      <td>1.300000</td>\n",
       "      <td>1.900000</td>\n",
       "      <td>76.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14643 rows Ã— 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM1687570    GSM1687571   GSM1687572    GSM1687573  \\\n",
       "Gene                                                                \n",
       " ABCB4        259.400000    180.500000     4.200000     26.000000   \n",
       " ACOT1         54.300000    269.600000    35.900000     63.400000   \n",
       " ACSM2          8.300000      4.900000     0.600000      5.400000   \n",
       " ACSM3      13418.200000  17818.800000  4635.400000  13449.900000   \n",
       " ADAM21P        3.700000     54.000000     4.900000      5.900000   \n",
       "...                  ...           ...          ...           ...   \n",
       "ZYG11BL        16.200000     40.233333    21.966667     42.900000   \n",
       "ZYX            54.400000     65.550000   123.800000    112.850000   \n",
       "ZZEF1          55.566667     98.433333    22.566667     44.566667   \n",
       "ZZZ3          231.000000    420.700000   415.200000    190.400000   \n",
       "dJ222E13.2     39.600000     40.000000    33.300000      2.400000   \n",
       "\n",
       "              GSM1687574    GSM1687575    GSM1687576  GSM1687577   GSM1687578  \\\n",
       "Gene                                                                            \n",
       " ABCB4         36.500000     50.600000     70.700000       17.90   166.100000   \n",
       " ACOT1         57.500000    323.600000     53.000000       14.40    35.100000   \n",
       " ACSM2          5.700000      2.500000     13.800000        8.10   102.700000   \n",
       " ACSM3      12204.700000  16411.300000  16341.800000     4182.80  9947.700000   \n",
       " ADAM21P        3.700000     36.900000     31.100000       14.70    32.700000   \n",
       "...                  ...           ...           ...         ...          ...   \n",
       "ZYG11BL        30.533333     44.800000     90.466667       26.20    34.166667   \n",
       "ZYX           174.700000    546.100000    433.400000      109.05   678.500000   \n",
       "ZZEF1          61.533333    137.966667    150.300000       27.20    47.433333   \n",
       "ZZZ3          415.400000    589.200000   1376.500000      260.10   251.500000   \n",
       "dJ222E13.2     32.500000     94.500000     86.400000       30.30     5.700000   \n",
       "\n",
       "             GSM1687579  ...   GSM1688358    GSM1688359   GSM1688360  \\\n",
       "Gene                     ...                                           \n",
       " ABCB4         6.800000  ...    13.300000     21.400000    14.500000   \n",
       " ACOT1        43.200000  ...    39.600000    124.400000    31.500000   \n",
       " ACSM2         3.000000  ...     3.000000      0.900000     0.800000   \n",
       " ACSM3      3564.800000  ...  9886.500000  10840.700000  3032.000000   \n",
       " ADAM21P      15.800000  ...    59.900000     15.500000     4.100000   \n",
       "...                 ...  ...          ...           ...          ...   \n",
       "ZYG11BL       22.400000  ...    44.466667     31.666667    11.966667   \n",
       "ZYX           80.300000  ...   488.650000     48.700000     8.500000   \n",
       "ZZEF1         48.366667  ...    50.700000     79.233333    12.933333   \n",
       "ZZZ3         331.100000  ...   257.500000    194.500000    60.000000   \n",
       "dJ222E13.2     4.600000  ...    35.500000     41.600000    14.300000   \n",
       "\n",
       "              GSM1688361   GSM1688362   GSM1688363    GSM1688364   GSM1688365  \\\n",
       "Gene                                                                            \n",
       " ABCB4        555.600000    13.600000     9.400000     26.300000     4.200000   \n",
       " ACOT1        201.800000    27.800000     9.500000    227.600000    21.600000   \n",
       " ACSM2          8.700000     1.900000     2.800000     19.000000     1.800000   \n",
       " ACSM3      10719.200000  8010.500000  3099.900000  13392.000000  3407.400000   \n",
       " ADAM21P       45.000000     3.200000     1.000000     17.600000    11.100000   \n",
       "...                  ...          ...          ...           ...          ...   \n",
       "ZYG11BL        19.133333    21.366667    11.933333     67.400000    20.833333   \n",
       "ZYX           125.550000   259.250000    58.500000    318.050000   384.800000   \n",
       "ZZEF1          74.133333    36.433333    14.100000     51.133333    16.133333   \n",
       "ZZZ3          236.600000   120.500000    43.600000    465.400000   139.400000   \n",
       "dJ222E13.2     48.500000    17.700000     1.800000     59.500000     1.300000   \n",
       "\n",
       "             GSM1688366    GSM1688367  \n",
       "Gene                                   \n",
       " ABCB4        23.300000     13.600000  \n",
       " ACOT1        33.200000    441.600000  \n",
       " ACSM2        14.800000      1.000000  \n",
       " ACSM3      8757.700000  11117.600000  \n",
       " ADAM21P      19.700000     25.500000  \n",
       "...                 ...           ...  \n",
       "ZYG11BL       27.266667     19.433333  \n",
       "ZYX          198.850000     66.600000  \n",
       "ZZEF1         41.200000     37.966667  \n",
       "ZZZ3         180.400000    266.700000  \n",
       "dJ222E13.2     1.900000     76.000000  \n",
       "\n",
       "[14643 rows x 798 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1687570</th>\n",
       "      <th>GSM1687571</th>\n",
       "      <th>GSM1687572</th>\n",
       "      <th>GSM1687573</th>\n",
       "      <th>GSM1687574</th>\n",
       "      <th>GSM1687575</th>\n",
       "      <th>GSM1687576</th>\n",
       "      <th>GSM1687577</th>\n",
       "      <th>GSM1687578</th>\n",
       "      <th>GSM1687579</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1688358</th>\n",
       "      <th>GSM1688359</th>\n",
       "      <th>GSM1688360</th>\n",
       "      <th>GSM1688361</th>\n",
       "      <th>GSM1688362</th>\n",
       "      <th>GSM1688363</th>\n",
       "      <th>GSM1688364</th>\n",
       "      <th>GSM1688365</th>\n",
       "      <th>GSM1688366</th>\n",
       "      <th>GSM1688367</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>15.400000</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>2.100000</td>\n",
       "      <td>37.600000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>24.300000</td>\n",
       "      <td>12.2</td>\n",
       "      <td>20.90</td>\n",
       "      <td>5.100000</td>\n",
       "      <td>6.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.70</td>\n",
       "      <td>26.800000</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>893.100000</td>\n",
       "      <td>27.600000</td>\n",
       "      <td>0.6</td>\n",
       "      <td>19.600000</td>\n",
       "      <td>8.700000</td>\n",
       "      <td>13.20</td>\n",
       "      <td>26.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>1.900000</td>\n",
       "      <td>7.800000</td>\n",
       "      <td>20.400000</td>\n",
       "      <td>4.200000</td>\n",
       "      <td>27.100000</td>\n",
       "      <td>54.400000</td>\n",
       "      <td>18.3</td>\n",
       "      <td>2.80</td>\n",
       "      <td>26.400000</td>\n",
       "      <td>3.900000</td>\n",
       "      <td>...</td>\n",
       "      <td>8.40</td>\n",
       "      <td>23.000000</td>\n",
       "      <td>5.700000</td>\n",
       "      <td>30.600000</td>\n",
       "      <td>7.700000</td>\n",
       "      <td>1.4</td>\n",
       "      <td>44.200000</td>\n",
       "      <td>14.200000</td>\n",
       "      <td>11.40</td>\n",
       "      <td>14.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>2.100000</td>\n",
       "      <td>34.900000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>1.300000</td>\n",
       "      <td>16.600000</td>\n",
       "      <td>26.000000</td>\n",
       "      <td>22.6</td>\n",
       "      <td>6.40</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>22.800000</td>\n",
       "      <td>...</td>\n",
       "      <td>9.70</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.600000</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>3.200000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.800000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>7.70</td>\n",
       "      <td>3.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>40.400000</td>\n",
       "      <td>94.200000</td>\n",
       "      <td>30.200000</td>\n",
       "      <td>88.000000</td>\n",
       "      <td>78.700000</td>\n",
       "      <td>181.900000</td>\n",
       "      <td>211.5</td>\n",
       "      <td>30.70</td>\n",
       "      <td>177.800000</td>\n",
       "      <td>16.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>97.60</td>\n",
       "      <td>123.700000</td>\n",
       "      <td>31.500000</td>\n",
       "      <td>105.000000</td>\n",
       "      <td>84.200000</td>\n",
       "      <td>21.5</td>\n",
       "      <td>144.100000</td>\n",
       "      <td>31.100000</td>\n",
       "      <td>58.40</td>\n",
       "      <td>36.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>140.800000</td>\n",
       "      <td>91.200000</td>\n",
       "      <td>167.200000</td>\n",
       "      <td>595.600000</td>\n",
       "      <td>481.200000</td>\n",
       "      <td>745.800000</td>\n",
       "      <td>901.4</td>\n",
       "      <td>178.30</td>\n",
       "      <td>402.700000</td>\n",
       "      <td>82.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>164.20</td>\n",
       "      <td>384.000000</td>\n",
       "      <td>127.700000</td>\n",
       "      <td>265.100000</td>\n",
       "      <td>72.100000</td>\n",
       "      <td>10.6</td>\n",
       "      <td>548.100000</td>\n",
       "      <td>52.300000</td>\n",
       "      <td>115.90</td>\n",
       "      <td>527.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>42.800000</td>\n",
       "      <td>32.500000</td>\n",
       "      <td>2.200000</td>\n",
       "      <td>31.600000</td>\n",
       "      <td>35.200000</td>\n",
       "      <td>39.800000</td>\n",
       "      <td>29.6</td>\n",
       "      <td>1.10</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>25.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.90</td>\n",
       "      <td>1.600000</td>\n",
       "      <td>4.100000</td>\n",
       "      <td>9.200000</td>\n",
       "      <td>16.800000</td>\n",
       "      <td>7.3</td>\n",
       "      <td>19.500000</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>16.90</td>\n",
       "      <td>4.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>63.100000</td>\n",
       "      <td>111.400000</td>\n",
       "      <td>61.500000</td>\n",
       "      <td>8.900000</td>\n",
       "      <td>208.300000</td>\n",
       "      <td>327.000000</td>\n",
       "      <td>496.3</td>\n",
       "      <td>91.60</td>\n",
       "      <td>102.900000</td>\n",
       "      <td>24.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>110.10</td>\n",
       "      <td>86.600000</td>\n",
       "      <td>15.500000</td>\n",
       "      <td>68.800000</td>\n",
       "      <td>73.900000</td>\n",
       "      <td>19.5</td>\n",
       "      <td>268.200000</td>\n",
       "      <td>19.300000</td>\n",
       "      <td>41.40</td>\n",
       "      <td>97.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>54.400000</td>\n",
       "      <td>65.550000</td>\n",
       "      <td>123.800000</td>\n",
       "      <td>112.850000</td>\n",
       "      <td>174.700000</td>\n",
       "      <td>546.100000</td>\n",
       "      <td>433.4</td>\n",
       "      <td>109.05</td>\n",
       "      <td>678.500000</td>\n",
       "      <td>80.300000</td>\n",
       "      <td>...</td>\n",
       "      <td>488.65</td>\n",
       "      <td>48.700000</td>\n",
       "      <td>8.500000</td>\n",
       "      <td>125.550000</td>\n",
       "      <td>259.250000</td>\n",
       "      <td>58.5</td>\n",
       "      <td>318.050000</td>\n",
       "      <td>384.800000</td>\n",
       "      <td>198.85</td>\n",
       "      <td>66.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>55.566667</td>\n",
       "      <td>98.433333</td>\n",
       "      <td>22.566667</td>\n",
       "      <td>44.566667</td>\n",
       "      <td>61.533333</td>\n",
       "      <td>137.966667</td>\n",
       "      <td>150.3</td>\n",
       "      <td>27.20</td>\n",
       "      <td>47.433333</td>\n",
       "      <td>48.366667</td>\n",
       "      <td>...</td>\n",
       "      <td>50.70</td>\n",
       "      <td>79.233333</td>\n",
       "      <td>12.933333</td>\n",
       "      <td>74.133333</td>\n",
       "      <td>36.433333</td>\n",
       "      <td>14.1</td>\n",
       "      <td>51.133333</td>\n",
       "      <td>16.133333</td>\n",
       "      <td>41.20</td>\n",
       "      <td>37.966667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>231.000000</td>\n",
       "      <td>420.700000</td>\n",
       "      <td>415.200000</td>\n",
       "      <td>190.400000</td>\n",
       "      <td>415.400000</td>\n",
       "      <td>589.200000</td>\n",
       "      <td>1376.5</td>\n",
       "      <td>260.10</td>\n",
       "      <td>251.500000</td>\n",
       "      <td>331.100000</td>\n",
       "      <td>...</td>\n",
       "      <td>257.50</td>\n",
       "      <td>194.500000</td>\n",
       "      <td>60.000000</td>\n",
       "      <td>236.600000</td>\n",
       "      <td>120.500000</td>\n",
       "      <td>43.6</td>\n",
       "      <td>465.400000</td>\n",
       "      <td>139.400000</td>\n",
       "      <td>180.40</td>\n",
       "      <td>266.700000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10387 rows Ã— 798 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM1687570  GSM1687571  GSM1687572  GSM1687573  GSM1687574  \\\n",
       "A2M      15.400000    3.500000    2.100000   37.600000    4.000000   \n",
       "A4GALT    1.900000    7.800000   20.400000    4.200000   27.100000   \n",
       "A4GNT     2.100000   34.900000    0.400000    1.300000   16.600000   \n",
       "AAAS     40.400000   94.200000   30.200000   88.000000   78.700000   \n",
       "AACS    140.800000   91.200000  167.200000  595.600000  481.200000   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZXDB     42.800000   32.500000    2.200000   31.600000   35.200000   \n",
       "ZXDC     63.100000  111.400000   61.500000    8.900000  208.300000   \n",
       "ZYX      54.400000   65.550000  123.800000  112.850000  174.700000   \n",
       "ZZEF1    55.566667   98.433333   22.566667   44.566667   61.533333   \n",
       "ZZZ3    231.000000  420.700000  415.200000  190.400000  415.400000   \n",
       "\n",
       "        GSM1687575  GSM1687576  GSM1687577  GSM1687578  GSM1687579  ...  \\\n",
       "A2M      24.300000        12.2       20.90    5.100000    6.300000  ...   \n",
       "A4GALT   54.400000        18.3        2.80   26.400000    3.900000  ...   \n",
       "A4GNT    26.000000        22.6        6.40   10.500000   22.800000  ...   \n",
       "AAAS    181.900000       211.5       30.70  177.800000   16.300000  ...   \n",
       "AACS    745.800000       901.4      178.30  402.700000   82.000000  ...   \n",
       "...            ...         ...         ...         ...         ...  ...   \n",
       "ZXDB     39.800000        29.6        1.10    4.000000   25.300000  ...   \n",
       "ZXDC    327.000000       496.3       91.60  102.900000   24.300000  ...   \n",
       "ZYX     546.100000       433.4      109.05  678.500000   80.300000  ...   \n",
       "ZZEF1   137.966667       150.3       27.20   47.433333   48.366667  ...   \n",
       "ZZZ3    589.200000      1376.5      260.10  251.500000  331.100000  ...   \n",
       "\n",
       "        GSM1688358  GSM1688359  GSM1688360  GSM1688361  GSM1688362  \\\n",
       "A2M           3.70   26.800000    0.900000  893.100000   27.600000   \n",
       "A4GALT        8.40   23.000000    5.700000   30.600000    7.700000   \n",
       "A4GNT         9.70    1.000000    1.600000    7.500000    3.200000   \n",
       "AAAS         97.60  123.700000   31.500000  105.000000   84.200000   \n",
       "AACS        164.20  384.000000  127.700000  265.100000   72.100000   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZXDB          2.90    1.600000    4.100000    9.200000   16.800000   \n",
       "ZXDC        110.10   86.600000   15.500000   68.800000   73.900000   \n",
       "ZYX         488.65   48.700000    8.500000  125.550000  259.250000   \n",
       "ZZEF1        50.70   79.233333   12.933333   74.133333   36.433333   \n",
       "ZZZ3        257.50  194.500000   60.000000  236.600000  120.500000   \n",
       "\n",
       "        GSM1688363  GSM1688364  GSM1688365  GSM1688366  GSM1688367  \n",
       "A2M            0.6   19.600000    8.700000       13.20   26.300000  \n",
       "A4GALT         1.4   44.200000   14.200000       11.40   14.200000  \n",
       "A4GNT          1.0   22.800000    5.000000        7.70    3.700000  \n",
       "AAAS          21.5  144.100000   31.100000       58.40   36.100000  \n",
       "AACS          10.6  548.100000   52.300000      115.90  527.100000  \n",
       "...            ...         ...         ...         ...         ...  \n",
       "ZXDB           7.3   19.500000    3.500000       16.90    4.200000  \n",
       "ZXDC          19.5  268.200000   19.300000       41.40   97.200000  \n",
       "ZYX           58.5  318.050000  384.800000      198.85   66.600000  \n",
       "ZZEF1         14.1   51.133333   16.133333       41.20   37.966667  \n",
       "ZZZ3          43.6  465.400000  139.400000      180.40  266.700000  \n",
       "\n",
       "[10387 rows x 798 columns]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>AANAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN2</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1687570</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.4</td>\n",
       "      <td>1.9</td>\n",
       "      <td>2.1</td>\n",
       "      <td>40.4</td>\n",
       "      <td>140.8</td>\n",
       "      <td>4.8</td>\n",
       "      <td>9.400000</td>\n",
       "      <td>97.6</td>\n",
       "      <td>0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.2</td>\n",
       "      <td>42.1</td>\n",
       "      <td>196.4</td>\n",
       "      <td>287.2</td>\n",
       "      <td>1944.3</td>\n",
       "      <td>42.8</td>\n",
       "      <td>63.1</td>\n",
       "      <td>54.40</td>\n",
       "      <td>55.566667</td>\n",
       "      <td>231.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1687571</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>7.8</td>\n",
       "      <td>34.9</td>\n",
       "      <td>94.2</td>\n",
       "      <td>91.2</td>\n",
       "      <td>25.5</td>\n",
       "      <td>33.766667</td>\n",
       "      <td>149.2</td>\n",
       "      <td>3.5</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4</td>\n",
       "      <td>48.4</td>\n",
       "      <td>341.1</td>\n",
       "      <td>407.7</td>\n",
       "      <td>2480.6</td>\n",
       "      <td>32.5</td>\n",
       "      <td>111.4</td>\n",
       "      <td>65.55</td>\n",
       "      <td>98.433333</td>\n",
       "      <td>420.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1687572</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.1</td>\n",
       "      <td>20.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>30.2</td>\n",
       "      <td>167.2</td>\n",
       "      <td>5.9</td>\n",
       "      <td>2.200000</td>\n",
       "      <td>179.1</td>\n",
       "      <td>0.3</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>304.2</td>\n",
       "      <td>248.7</td>\n",
       "      <td>388.6</td>\n",
       "      <td>2.2</td>\n",
       "      <td>61.5</td>\n",
       "      <td>123.80</td>\n",
       "      <td>22.566667</td>\n",
       "      <td>415.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1687573</th>\n",
       "      <td>0.0</td>\n",
       "      <td>37.6</td>\n",
       "      <td>4.2</td>\n",
       "      <td>1.3</td>\n",
       "      <td>88.0</td>\n",
       "      <td>595.6</td>\n",
       "      <td>8.4</td>\n",
       "      <td>17.566667</td>\n",
       "      <td>203.7</td>\n",
       "      <td>2.2</td>\n",
       "      <td>...</td>\n",
       "      <td>14.1</td>\n",
       "      <td>1.3</td>\n",
       "      <td>140.5</td>\n",
       "      <td>271.4</td>\n",
       "      <td>1544.0</td>\n",
       "      <td>31.6</td>\n",
       "      <td>8.9</td>\n",
       "      <td>112.85</td>\n",
       "      <td>44.566667</td>\n",
       "      <td>190.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1687574</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>27.1</td>\n",
       "      <td>16.6</td>\n",
       "      <td>78.7</td>\n",
       "      <td>481.2</td>\n",
       "      <td>4.1</td>\n",
       "      <td>11.433333</td>\n",
       "      <td>684.4</td>\n",
       "      <td>7.8</td>\n",
       "      <td>...</td>\n",
       "      <td>14.5</td>\n",
       "      <td>31.0</td>\n",
       "      <td>282.9</td>\n",
       "      <td>253.0</td>\n",
       "      <td>3481.2</td>\n",
       "      <td>35.2</td>\n",
       "      <td>208.3</td>\n",
       "      <td>174.70</td>\n",
       "      <td>61.533333</td>\n",
       "      <td>415.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1688363</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>1.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21.5</td>\n",
       "      <td>10.6</td>\n",
       "      <td>2.7</td>\n",
       "      <td>3.233333</td>\n",
       "      <td>74.9</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8</td>\n",
       "      <td>1.5</td>\n",
       "      <td>34.0</td>\n",
       "      <td>76.4</td>\n",
       "      <td>1042.9</td>\n",
       "      <td>7.3</td>\n",
       "      <td>19.5</td>\n",
       "      <td>58.50</td>\n",
       "      <td>14.100000</td>\n",
       "      <td>43.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1688364</th>\n",
       "      <td>0.0</td>\n",
       "      <td>19.6</td>\n",
       "      <td>44.2</td>\n",
       "      <td>22.8</td>\n",
       "      <td>144.1</td>\n",
       "      <td>548.1</td>\n",
       "      <td>142.3</td>\n",
       "      <td>6.133333</td>\n",
       "      <td>975.1</td>\n",
       "      <td>12.2</td>\n",
       "      <td>...</td>\n",
       "      <td>25.6</td>\n",
       "      <td>2.4</td>\n",
       "      <td>341.7</td>\n",
       "      <td>396.6</td>\n",
       "      <td>2340.7</td>\n",
       "      <td>19.5</td>\n",
       "      <td>268.2</td>\n",
       "      <td>318.05</td>\n",
       "      <td>51.133333</td>\n",
       "      <td>465.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1688365</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.7</td>\n",
       "      <td>14.2</td>\n",
       "      <td>5.0</td>\n",
       "      <td>31.1</td>\n",
       "      <td>52.3</td>\n",
       "      <td>2.2</td>\n",
       "      <td>2.733333</td>\n",
       "      <td>98.8</td>\n",
       "      <td>1.3</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.1</td>\n",
       "      <td>47.5</td>\n",
       "      <td>34.1</td>\n",
       "      <td>91.2</td>\n",
       "      <td>3.5</td>\n",
       "      <td>19.3</td>\n",
       "      <td>384.80</td>\n",
       "      <td>16.133333</td>\n",
       "      <td>139.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1688366</th>\n",
       "      <td>0.0</td>\n",
       "      <td>13.2</td>\n",
       "      <td>11.4</td>\n",
       "      <td>7.7</td>\n",
       "      <td>58.4</td>\n",
       "      <td>115.9</td>\n",
       "      <td>6.4</td>\n",
       "      <td>8.133333</td>\n",
       "      <td>175.9</td>\n",
       "      <td>0.6</td>\n",
       "      <td>...</td>\n",
       "      <td>4.9</td>\n",
       "      <td>3.7</td>\n",
       "      <td>110.5</td>\n",
       "      <td>472.5</td>\n",
       "      <td>2737.3</td>\n",
       "      <td>16.9</td>\n",
       "      <td>41.4</td>\n",
       "      <td>198.85</td>\n",
       "      <td>41.200000</td>\n",
       "      <td>180.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1688367</th>\n",
       "      <td>0.0</td>\n",
       "      <td>26.3</td>\n",
       "      <td>14.2</td>\n",
       "      <td>3.7</td>\n",
       "      <td>36.1</td>\n",
       "      <td>527.1</td>\n",
       "      <td>29.5</td>\n",
       "      <td>14.100000</td>\n",
       "      <td>261.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4</td>\n",
       "      <td>48.1</td>\n",
       "      <td>320.1</td>\n",
       "      <td>304.1</td>\n",
       "      <td>1168.5</td>\n",
       "      <td>4.2</td>\n",
       "      <td>97.2</td>\n",
       "      <td>66.60</td>\n",
       "      <td>37.966667</td>\n",
       "      <td>266.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>798 rows Ã— 10388 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Adrenocortical Cancer   A2M  A4GALT  A4GNT   AAAS   AACS  AADAC  \\\n",
       "GSM1687570                    0.0  15.4     1.9    2.1   40.4  140.8    4.8   \n",
       "GSM1687571                    0.0   3.5     7.8   34.9   94.2   91.2   25.5   \n",
       "GSM1687572                    0.0   2.1    20.4    0.4   30.2  167.2    5.9   \n",
       "GSM1687573                    0.0  37.6     4.2    1.3   88.0  595.6    8.4   \n",
       "GSM1687574                    0.0   4.0    27.1   16.6   78.7  481.2    4.1   \n",
       "...                           ...   ...     ...    ...    ...    ...    ...   \n",
       "GSM1688363                    0.0   0.6     1.4    1.0   21.5   10.6    2.7   \n",
       "GSM1688364                    0.0  19.6    44.2   22.8  144.1  548.1  142.3   \n",
       "GSM1688365                    0.0   8.7    14.2    5.0   31.1   52.3    2.2   \n",
       "GSM1688366                    0.0  13.2    11.4    7.7   58.4  115.9    6.4   \n",
       "GSM1688367                    0.0  26.3    14.2    3.7   36.1  527.1   29.5   \n",
       "\n",
       "                 AAK1   AAMP  AANAT  ...  ZSCAN2  ZSWIM1   ZW10  ZWILCH  \\\n",
       "GSM1687570   9.400000   97.6    0.5  ...     2.2    42.1  196.4   287.2   \n",
       "GSM1687571  33.766667  149.2    3.5  ...     2.4    48.4  341.1   407.7   \n",
       "GSM1687572   2.200000  179.1    0.3  ...     1.2     6.0  304.2   248.7   \n",
       "GSM1687573  17.566667  203.7    2.2  ...    14.1     1.3  140.5   271.4   \n",
       "GSM1687574  11.433333  684.4    7.8  ...    14.5    31.0  282.9   253.0   \n",
       "...               ...    ...    ...  ...     ...     ...    ...     ...   \n",
       "GSM1688363   3.233333   74.9    0.4  ...     0.8     1.5   34.0    76.4   \n",
       "GSM1688364   6.133333  975.1   12.2  ...    25.6     2.4  341.7   396.6   \n",
       "GSM1688365   2.733333   98.8    1.3  ...     1.0     3.1   47.5    34.1   \n",
       "GSM1688366   8.133333  175.9    0.6  ...     4.9     3.7  110.5   472.5   \n",
       "GSM1688367  14.100000  261.2    1.0  ...     2.4    48.1  320.1   304.1   \n",
       "\n",
       "             ZWINT  ZXDB   ZXDC     ZYX      ZZEF1   ZZZ3  \n",
       "GSM1687570  1944.3  42.8   63.1   54.40  55.566667  231.0  \n",
       "GSM1687571  2480.6  32.5  111.4   65.55  98.433333  420.7  \n",
       "GSM1687572   388.6   2.2   61.5  123.80  22.566667  415.2  \n",
       "GSM1687573  1544.0  31.6    8.9  112.85  44.566667  190.4  \n",
       "GSM1687574  3481.2  35.2  208.3  174.70  61.533333  415.4  \n",
       "...            ...   ...    ...     ...        ...    ...  \n",
       "GSM1688363  1042.9   7.3   19.5   58.50  14.100000   43.6  \n",
       "GSM1688364  2340.7  19.5  268.2  318.05  51.133333  465.4  \n",
       "GSM1688365    91.2   3.5   19.3  384.80  16.133333  139.4  \n",
       "GSM1688366  2737.3  16.9   41.4  198.85  41.200000  180.4  \n",
       "GSM1688367  1168.5   4.2   97.2   66.60  37.966667  266.7  \n",
       "\n",
       "[798 rows x 10388 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 798 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 2 occurrences. This represents 0.25% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Adrenocortical Carcinoma Gene Expression Profiling [Affymetrix]\"\n",
      "!Series_summary\t\"Background: Adrenocortical carcinoma (ACC) is associated with poor survival rates.  The objective of the study was to analyze ACC gene expression profiling data prognostic biomarkers and novel therapeutic targets.\"\n",
      "!Series_summary\t\"Methods: 44 ACC and 4 normal adrenal glands were profiled on Affymetrix U133 Plus 2 expression microarrays and pathway and transcriptional enrichment analysis performed.  Protein levels were determined by western blot.  Drug efficacy was assessed against ACC cell lines.  Previously published expression datasets were analyzed as validation data sets.\"\n",
      "!Series_summary\t\"Results: Pathway enrichment analysis identified marked dysregulation of cyclin-dependent kinases and mitosis.   Over-expression of PTTG1, which encodes securin, a negative regulator of p53, was identified as a marker of poor survival.  Median survival for patients with tumors expressing high PTTG1 levels (log2 ratio of PTTG1 to average beta-actin <-3.04 ) was 1.8 years compared to 9.0 years if tumors expressed lower levels of PTTG1 (P<0.0001).  These findings were confirmed by our analysis of previously published datasets.  Treatment of ACC cell lines with vorinostat decreased securin levels and inhibited cell growth (IC50s of 1.69 uM and 0.891 uM, for SW-13 and H295R, respectively).\"\n",
      "!Series_summary\t\"Conclusion: Over-expression of PTTG1 is correlated with poor survival in ACC.  PTTG1/securin is a prognostic biomarker and warrants investigation as a therapeutic target.\"\n",
      "!Series_overall_design\t\"RNA from forty-four adrenocortical carcinomas and four normal adrenal glands was extracted, labeled, and hybridized to Affymetrix U133 Plus 2 arrays.  The resulting data was normalized by gcRMA with quantile normalization and background subtraction after using the ExpressionFileCreator in GenePattern.  Data was then floored at 5.5 using PreprocessDataset, and filtered to remove 1) probes with more than 35 floored values and/or 2) probes where all values from one batch were floored while values from the other batch were not.  Further batch effects were minimized using ComBat with the parametric option.   Data was then floored at 2.  Differentially expressed genes were determined using a T-test with multiple comparison correction as implemented by Comparative Marker Selection in Gene Pattern.  Genes with the corrected p-value < 0.005 and the FDR < 0.075 were selected for further study.  For comparing high to low grade or primary to recurrence, the FDR cut-off was increased to < 0.13.  Survival analysis was conducted using Prism 6 (GraphPad) to generate Kaplan-Meier curves that were compared by log-rank.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: NA</td>\n",
       "      <td>Stage: 2</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 2</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>...</td>\n",
       "      <td>Stage: 1</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 2</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: Recurrence</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: 4</td>\n",
       "      <td>Stage: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: NA</td>\n",
       "      <td>tumor grade: 3</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor grade: 1</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: 2</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: 4</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "      <td>tumor grade: Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: NA</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: Cushings</td>\n",
       "      <td>functional: Cushings</td>\n",
       "      <td>...</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: Unknown</td>\n",
       "      <td>functional: Cortisol, aldosterone, testosterone</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: aldosterone</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: None</td>\n",
       "      <td>functional: Unknown</td>\n",
       "      <td>functional: Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: Unknown</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: Unknown</td>\n",
       "      <td>age in years: 23.3</td>\n",
       "      <td>age in years: 56.5</td>\n",
       "      <td>age in years: 67.8</td>\n",
       "      <td>age in years: 72.1</td>\n",
       "      <td>age in years: 46.9</td>\n",
       "      <td>...</td>\n",
       "      <td>age in years: 57</td>\n",
       "      <td>age in years: 59</td>\n",
       "      <td>age in years: 59</td>\n",
       "      <td>age in years: 55</td>\n",
       "      <td>age in years: 51</td>\n",
       "      <td>age in years: 53</td>\n",
       "      <td>age in years: 69</td>\n",
       "      <td>age in years: 63</td>\n",
       "      <td>age in years: 28</td>\n",
       "      <td>age in years: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: NA</td>\n",
       "      <td>survival in years: 3</td>\n",
       "      <td>survival in years: 0.6</td>\n",
       "      <td>survival in years: 1.7</td>\n",
       "      <td>survival in years: 0.4</td>\n",
       "      <td>survival in years: 0.1</td>\n",
       "      <td>...</td>\n",
       "      <td>survival in years: 3</td>\n",
       "      <td>survival in years: 7.583</td>\n",
       "      <td>survival in years: Unknown</td>\n",
       "      <td>survival in years: 0.583</td>\n",
       "      <td>survival in years: 6</td>\n",
       "      <td>survival in years: 2.083</td>\n",
       "      <td>survival in years: 2.83</td>\n",
       "      <td>survival in years: 2.08</td>\n",
       "      <td>survival in years: Unknown</td>\n",
       "      <td>survival in years: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: NA</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>...</td>\n",
       "      <td>survival status: alive</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: Unknown</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: alive</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: dead</td>\n",
       "      <td>survival status: alive</td>\n",
       "      <td>survival status: Unknown</td>\n",
       "      <td>survival status: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: NA</td>\n",
       "      <td>tumor size in cm: 19</td>\n",
       "      <td>tumor size in cm: 9</td>\n",
       "      <td>tumor size in cm: 7.6</td>\n",
       "      <td>tumor size in cm: 9.5</td>\n",
       "      <td>tumor size in cm: 12</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor size in cm: 4</td>\n",
       "      <td>tumor size in cm: 2.5</td>\n",
       "      <td>tumor size in cm: 10</td>\n",
       "      <td>tumor size in cm: 10.5</td>\n",
       "      <td>tumor size in cm: 14.5</td>\n",
       "      <td>tumor size in cm: 14.5</td>\n",
       "      <td>tumor size in cm: 7.8</td>\n",
       "      <td>tumor size in cm: 7.8</td>\n",
       "      <td>tumor size in cm: Unknown</td>\n",
       "      <td>tumor size in cm: Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: NA</td>\n",
       "      <td>tumor weight in grams: 1100</td>\n",
       "      <td>tumor weight in grams: 190</td>\n",
       "      <td>tumor weight in grams: 150</td>\n",
       "      <td>tumor weight in grams: 175</td>\n",
       "      <td>tumor weight in grams: 235</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor weight in grams: 39</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "      <td>tumor weight in grams: 22</td>\n",
       "      <td>tumor weight in grams: 277</td>\n",
       "      <td>tumor weight in grams: 325</td>\n",
       "      <td>tumor weight in grams: 1243</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "      <td>tumor weight in grams: 132</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "      <td>tumor weight in grams: unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                  GSM493251  \\\n",
       "0  !Sample_characteristics_ch1                  Stage: NA   \n",
       "1  !Sample_characteristics_ch1            tumor grade: NA   \n",
       "2  !Sample_characteristics_ch1             functional: NA   \n",
       "3  !Sample_characteristics_ch1            gender: Unknown   \n",
       "4  !Sample_characteristics_ch1      age in years: Unknown   \n",
       "5  !Sample_characteristics_ch1      survival in years: NA   \n",
       "6  !Sample_characteristics_ch1        survival status: NA   \n",
       "7  !Sample_characteristics_ch1       tumor size in cm: NA   \n",
       "8  !Sample_characteristics_ch1  tumor weight in grams: NA   \n",
       "9  !Sample_characteristics_ch1                   batch: 1   \n",
       "\n",
       "                   GSM493252                  GSM493253  \\\n",
       "0                  Stage: NA                  Stage: NA   \n",
       "1            tumor grade: NA            tumor grade: NA   \n",
       "2             functional: NA             functional: NA   \n",
       "3            gender: Unknown            gender: Unknown   \n",
       "4      age in years: Unknown      age in years: Unknown   \n",
       "5      survival in years: NA      survival in years: NA   \n",
       "6        survival status: NA        survival status: NA   \n",
       "7       tumor size in cm: NA       tumor size in cm: NA   \n",
       "8  tumor weight in grams: NA  tumor weight in grams: NA   \n",
       "9                   batch: 1                   batch: 1   \n",
       "\n",
       "                   GSM493254                    GSM493255  \\\n",
       "0                  Stage: NA                     Stage: 2   \n",
       "1            tumor grade: NA               tumor grade: 3   \n",
       "2             functional: NA             functional: None   \n",
       "3            gender: Unknown                    gender: M   \n",
       "4      age in years: Unknown           age in years: 23.3   \n",
       "5      survival in years: NA         survival in years: 3   \n",
       "6        survival status: NA        survival status: dead   \n",
       "7       tumor size in cm: NA         tumor size in cm: 19   \n",
       "8  tumor weight in grams: NA  tumor weight in grams: 1100   \n",
       "9                   batch: 1                     batch: 1   \n",
       "\n",
       "                    GSM493256                   GSM493257  \\\n",
       "0                    Stage: 4                    Stage: 2   \n",
       "1              tumor grade: 4              tumor grade: 4   \n",
       "2            functional: None            functional: None   \n",
       "3                   gender: F                   gender: M   \n",
       "4          age in years: 56.5          age in years: 67.8   \n",
       "5      survival in years: 0.6      survival in years: 1.7   \n",
       "6       survival status: dead       survival status: dead   \n",
       "7         tumor size in cm: 9       tumor size in cm: 7.6   \n",
       "8  tumor weight in grams: 190  tumor weight in grams: 150   \n",
       "9                    batch: 1                    batch: 1   \n",
       "\n",
       "                    GSM493258                   GSM493259  ...  \\\n",
       "0                    Stage: 4                    Stage: 4  ...   \n",
       "1              tumor grade: 4              tumor grade: 4  ...   \n",
       "2        functional: Cushings        functional: Cushings  ...   \n",
       "3                   gender: M                   gender: M  ...   \n",
       "4          age in years: 72.1          age in years: 46.9  ...   \n",
       "5      survival in years: 0.4      survival in years: 0.1  ...   \n",
       "6       survival status: dead       survival status: dead  ...   \n",
       "7       tumor size in cm: 9.5        tumor size in cm: 12  ...   \n",
       "8  tumor weight in grams: 175  tumor weight in grams: 235  ...   \n",
       "9                    batch: 1                    batch: 1  ...   \n",
       "\n",
       "                  GSM1094071                      GSM1094072  \\\n",
       "0                   Stage: 1               Stage: Recurrence   \n",
       "1             tumor grade: 1            tumor grade: Unknown   \n",
       "2           functional: None                functional: None   \n",
       "3                  gender: M                       gender: M   \n",
       "4           age in years: 57                age in years: 59   \n",
       "5       survival in years: 3        survival in years: 7.583   \n",
       "6     survival status: alive           survival status: dead   \n",
       "7        tumor size in cm: 4           tumor size in cm: 2.5   \n",
       "8  tumor weight in grams: 39  tumor weight in grams: unknown   \n",
       "9                   batch: 2                        batch: 2   \n",
       "\n",
       "                   GSM1094073  \\\n",
       "0                    Stage: 4   \n",
       "1        tumor grade: Unknown   \n",
       "2         functional: Unknown   \n",
       "3                   gender: F   \n",
       "4            age in years: 59   \n",
       "5  survival in years: Unknown   \n",
       "6    survival status: Unknown   \n",
       "7        tumor size in cm: 10   \n",
       "8   tumor weight in grams: 22   \n",
       "9                    batch: 2   \n",
       "\n",
       "                                        GSM1094074  \\\n",
       "0                                         Stage: 2   \n",
       "1                                   tumor grade: 4   \n",
       "2  functional: Cortisol, aldosterone, testosterone   \n",
       "3                                        gender: F   \n",
       "4                                 age in years: 55   \n",
       "5                         survival in years: 0.583   \n",
       "6                            survival status: dead   \n",
       "7                           tumor size in cm: 10.5   \n",
       "8                       tumor weight in grams: 277   \n",
       "9                                         batch: 2   \n",
       "\n",
       "                   GSM1094075                   GSM1094076  \\\n",
       "0           Stage: Recurrence            Stage: Recurrence   \n",
       "1              tumor grade: 2         tumor grade: Unknown   \n",
       "2            functional: None      functional: aldosterone   \n",
       "3                   gender: F                    gender: F   \n",
       "4            age in years: 51             age in years: 53   \n",
       "5        survival in years: 6     survival in years: 2.083   \n",
       "6      survival status: alive        survival status: dead   \n",
       "7      tumor size in cm: 14.5       tumor size in cm: 14.5   \n",
       "8  tumor weight in grams: 325  tumor weight in grams: 1243   \n",
       "9                    batch: 2                     batch: 2   \n",
       "\n",
       "                       GSM1094077                  GSM1094078  \\\n",
       "0               Stage: Recurrence                    Stage: 4   \n",
       "1            tumor grade: Unknown              tumor grade: 4   \n",
       "2                functional: None            functional: None   \n",
       "3                       gender: M                   gender: M   \n",
       "4                age in years: 69            age in years: 63   \n",
       "5         survival in years: 2.83     survival in years: 2.08   \n",
       "6           survival status: dead      survival status: alive   \n",
       "7           tumor size in cm: 7.8       tumor size in cm: 7.8   \n",
       "8  tumor weight in grams: unknown  tumor weight in grams: 132   \n",
       "9                        batch: 2                    batch: 2   \n",
       "\n",
       "                       GSM1094079                      GSM1094080  \n",
       "0                        Stage: 4                       Stage: NA  \n",
       "1            tumor grade: Unknown            tumor grade: Unknown  \n",
       "2             functional: Unknown             functional: Unknown  \n",
       "3                       gender: F                      gender: NA  \n",
       "4                age in years: 28                age in years: NA  \n",
       "5      survival in years: Unknown           survival in years: NA  \n",
       "6        survival status: Unknown             survival status: NA  \n",
       "7       tumor size in cm: Unknown       tumor size in cm: Unknown  \n",
       "8  tumor weight in grams: unknown  tumor weight in grams: unknown  \n",
       "9                        batch: 2                        batch: 2  \n",
       "\n",
       "[10 rows x 49 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished (tumor_gradeé—®äº†TA)\n",
    "\n",
    "cohort = accession_num = \"GSE19750\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'tumor grade: NA', 'tumor grade: 3',\n",
       "       'tumor grade: 4', 'tumor grade: 2', 'tumor grade: 1',\n",
       "       'tumor grade: Unknown'], dtype=object)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1\n",
       "tumor grade: Unknown           16\n",
       "tumor grade: 2                  9\n",
       "tumor grade: 4                  8\n",
       "tumor grade: 1                  7\n",
       "tumor grade: NA                 4\n",
       "tumor grade: 3                  4\n",
       "!Sample_characteristics_ch1     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_counts = tumor_stage_row.value_counts()\n",
    "tumor_stage_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 4\n",
    "gender_row = 3\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "def convert_trait(tumor_grade):\n",
    "    if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "        return 1  \n",
    "    elif tumor_grade == 'tumor grade: 1':\n",
    "        return 0  \n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>57</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>55</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "      <td>69</td>\n",
       "      <td>63</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      GSM493251 GSM493252 GSM493253 GSM493254 GSM493255  \\\n",
       "Adrenocortical Cancer      None      None      None      None         1   \n",
       "Age                        None      None      None      None      None   \n",
       "Gender                     None      None      None      None         0   \n",
       "\n",
       "                      GSM493256 GSM493257 GSM493258 GSM493259 GSM493260  ...  \\\n",
       "Adrenocortical Cancer         1         1         1         1         1  ...   \n",
       "Age                        None      None      None      None      None  ...   \n",
       "Gender                        1         0         0         0         1  ...   \n",
       "\n",
       "                      GSM1094071 GSM1094072 GSM1094073 GSM1094074  GSM1094075  \\\n",
       "Adrenocortical Cancer          0       None       None          1           1   \n",
       "Age                           57         59         59         55          51   \n",
       "Gender                         0          0          1          1           1   \n",
       "\n",
       "                      GSM1094076 GSM1094077 GSM1094078 GSM1094079  GSM1094080  \n",
       "Adrenocortical Cancer       None       None          1       None        None  \n",
       "Age                           53         69         63         28        None  \n",
       "Gender                         1          0          0          1        None  \n",
       "\n",
       "[3 rows x 48 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>101.10</td>\n",
       "      <td>48.86</td>\n",
       "      <td>100.83</td>\n",
       "      <td>104.84</td>\n",
       "      <td>566.78</td>\n",
       "      <td>418.37</td>\n",
       "      <td>499.31</td>\n",
       "      <td>432.83</td>\n",
       "      <td>212.42</td>\n",
       "      <td>336.85</td>\n",
       "      <td>...</td>\n",
       "      <td>139.79</td>\n",
       "      <td>35.72</td>\n",
       "      <td>79.18</td>\n",
       "      <td>154.74</td>\n",
       "      <td>166.63</td>\n",
       "      <td>29.67</td>\n",
       "      <td>72.48</td>\n",
       "      <td>10.21</td>\n",
       "      <td>151.86</td>\n",
       "      <td>137.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>22.58</td>\n",
       "      <td>18.30</td>\n",
       "      <td>16.96</td>\n",
       "      <td>16.96</td>\n",
       "      <td>18.64</td>\n",
       "      <td>39.96</td>\n",
       "      <td>57.40</td>\n",
       "      <td>34.15</td>\n",
       "      <td>43.71</td>\n",
       "      <td>32.52</td>\n",
       "      <td>...</td>\n",
       "      <td>8.32</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.64</td>\n",
       "      <td>17.01</td>\n",
       "      <td>17.19</td>\n",
       "      <td>25.84</td>\n",
       "      <td>16.13</td>\n",
       "      <td>43.64</td>\n",
       "      <td>11.46</td>\n",
       "      <td>22.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>73.33</td>\n",
       "      <td>30.19</td>\n",
       "      <td>155.69</td>\n",
       "      <td>173.28</td>\n",
       "      <td>14.43</td>\n",
       "      <td>224.29</td>\n",
       "      <td>30.10</td>\n",
       "      <td>9.52</td>\n",
       "      <td>9.17</td>\n",
       "      <td>9.09</td>\n",
       "      <td>...</td>\n",
       "      <td>24.91</td>\n",
       "      <td>11.09</td>\n",
       "      <td>309.80</td>\n",
       "      <td>42.61</td>\n",
       "      <td>9.57</td>\n",
       "      <td>38.21</td>\n",
       "      <td>50.33</td>\n",
       "      <td>321.67</td>\n",
       "      <td>14.87</td>\n",
       "      <td>9.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>11.97</td>\n",
       "      <td>10.74</td>\n",
       "      <td>9.91</td>\n",
       "      <td>9.77</td>\n",
       "      <td>8.73</td>\n",
       "      <td>13.49</td>\n",
       "      <td>8.65</td>\n",
       "      <td>9.59</td>\n",
       "      <td>12.71</td>\n",
       "      <td>8.71</td>\n",
       "      <td>...</td>\n",
       "      <td>8.73</td>\n",
       "      <td>18.09</td>\n",
       "      <td>9.59</td>\n",
       "      <td>8.69</td>\n",
       "      <td>8.69</td>\n",
       "      <td>12.21</td>\n",
       "      <td>8.64</td>\n",
       "      <td>8.64</td>\n",
       "      <td>8.68</td>\n",
       "      <td>8.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.76</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>...</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM493251  GSM493252  GSM493253  GSM493254  GSM493255  \\\n",
       "ID                                                                       \n",
       "1007_s_at           101.10      48.86     100.83     104.84     566.78   \n",
       "1053_at              22.58      18.30      16.96      16.96      18.64   \n",
       "117_at               73.33      30.19     155.69     173.28      14.43   \n",
       "121_at               11.97      10.74       9.91       9.77       8.73   \n",
       "1255_g_at             5.50       5.50       5.50       5.50       5.50   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "AFFX-ThrX-5_at        5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-ThrX-M_at        5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-TrpnX-3_at       5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-TrpnX-5_at       5.50       5.50       5.50       5.50       5.50   \n",
       "AFFX-TrpnX-M_at       5.50       5.50       5.50       5.76       5.50   \n",
       "\n",
       "                 GSM493256  GSM493257  GSM493258  GSM493259  GSM493260  ...  \\\n",
       "ID                                                                      ...   \n",
       "1007_s_at           418.37     499.31     432.83     212.42     336.85  ...   \n",
       "1053_at              39.96      57.40      34.15      43.71      32.52  ...   \n",
       "117_at              224.29      30.10       9.52       9.17       9.09  ...   \n",
       "121_at               13.49       8.65       9.59      12.71       8.71  ...   \n",
       "1255_g_at             5.50       5.50       5.50       5.50       5.50  ...   \n",
       "...                    ...        ...        ...        ...        ...  ...   \n",
       "AFFX-ThrX-5_at        5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-ThrX-M_at        5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-TrpnX-3_at       5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-TrpnX-5_at       5.50       5.50       5.50       5.50       5.50  ...   \n",
       "AFFX-TrpnX-M_at       5.50       5.50       5.50       5.50       5.50  ...   \n",
       "\n",
       "                 GSM1094071  GSM1094072  GSM1094073  GSM1094074  GSM1094075  \\\n",
       "ID                                                                            \n",
       "1007_s_at            139.79       35.72       79.18      154.74      166.63   \n",
       "1053_at                8.32        8.79       11.64       17.01       17.19   \n",
       "117_at                24.91       11.09      309.80       42.61        9.57   \n",
       "121_at                 8.73       18.09        9.59        8.69        8.69   \n",
       "1255_g_at              5.50        5.50        5.50        5.50        5.50   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at         5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-ThrX-M_at         5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-TrpnX-3_at        5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-TrpnX-5_at        5.50        5.50        5.50        5.50        5.50   \n",
       "AFFX-TrpnX-M_at        5.50        5.50        5.50        5.50        5.50   \n",
       "\n",
       "                 GSM1094076  GSM1094077  GSM1094078  GSM1094079  GSM1094080  \n",
       "ID                                                                           \n",
       "1007_s_at             29.67       72.48       10.21      151.86      137.73  \n",
       "1053_at               25.84       16.13       43.64       11.46       22.56  \n",
       "117_at                38.21       50.33      321.67       14.87        9.55  \n",
       "121_at                12.21        8.64        8.64        8.68        8.69  \n",
       "1255_g_at              5.50        5.50        5.50        5.50        5.50  \n",
       "...                     ...         ...         ...         ...         ...  \n",
       "AFFX-ThrX-5_at         5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-ThrX-M_at         5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-TrpnX-3_at        5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-TrpnX-5_at        5.50        5.50        5.50        5.50        5.50  \n",
       "AFFX-TrpnX-M_at        5.50        5.50        5.50        5.50        5.50  \n",
       "\n",
       "[54675 rows x 48 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679118</th>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679119</th>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679120</th>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679121</th>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679122</th>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>5.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2679123 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID  GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0              1007_s_at  U48705     NaN            Homo sapiens   \n",
       "1                1053_at  M87338     NaN            Homo sapiens   \n",
       "2                 117_at  X51757     NaN            Homo sapiens   \n",
       "3                 121_at  X69699     NaN            Homo sapiens   \n",
       "4              1255_g_at  L36861     NaN            Homo sapiens   \n",
       "...                  ...     ...     ...                     ...   \n",
       "2679118   AFFX-ThrX-5_at     5.5     NaN                     NaN   \n",
       "2679119   AFFX-ThrX-M_at     5.5     NaN                     NaN   \n",
       "2679120  AFFX-TrpnX-3_at     5.5     NaN                     NaN   \n",
       "2679121  AFFX-TrpnX-5_at     5.5     NaN                     NaN   \n",
       "2679122  AFFX-TrpnX-M_at     5.5     NaN                     NaN   \n",
       "\n",
       "        Annotation Date      Sequence Type                  Sequence Source  \\\n",
       "0           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "2           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "4           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                 ...                ...                              ...   \n",
       "2679118             NaN                NaN                              NaN   \n",
       "2679119             NaN                NaN                              NaN   \n",
       "2679120             NaN                NaN                              NaN   \n",
       "2679121             NaN                NaN                              NaN   \n",
       "2679122             NaN                NaN                              NaN   \n",
       "\n",
       "                                        Target Description  \\\n",
       "0        U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1        M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2        X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3        X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4        L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                    ...   \n",
       "2679118                                                NaN   \n",
       "2679119                                                NaN   \n",
       "2679120                                                NaN   \n",
       "2679121                                                NaN   \n",
       "2679122                                                NaN   \n",
       "\n",
       "        Representative Public ID  \\\n",
       "0                         U48705   \n",
       "1                         M87338   \n",
       "2                         X51757   \n",
       "3                         X69699   \n",
       "4                         L36861   \n",
       "...                          ...   \n",
       "2679118                      NaN   \n",
       "2679119                      NaN   \n",
       "2679120                      NaN   \n",
       "2679121                      NaN   \n",
       "2679122                      NaN   \n",
       "\n",
       "                                                Gene Title       Gene Symbol  \\\n",
       "0        discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1              replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2                     heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                             paired box 8              PAX8   \n",
       "4                  guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "...                                                    ...               ...   \n",
       "2679118                                                NaN               NaN   \n",
       "2679119                                                NaN               NaN   \n",
       "2679120                                                NaN               NaN   \n",
       "2679121                                                NaN               NaN   \n",
       "2679122                                                NaN               NaN   \n",
       "\n",
       "            ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0        780 /// 100616237  NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1                     5982  NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2                     3310                                          NM_002155   \n",
       "3                     7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                     2978                         NM_000409 /// XM_006715073   \n",
       "...                    ...                                                ...   \n",
       "2679118                NaN                                                NaN   \n",
       "2679119                NaN                                                NaN   \n",
       "2679120                NaN                                                NaN   \n",
       "2679121                NaN                                                NaN   \n",
       "2679122                NaN                                                NaN   \n",
       "\n",
       "                          Gene Ontology Biological Process  \\\n",
       "0        0001558 // regulation of cell growth // inferr...   \n",
       "1        0000278 // mitotic cell cycle // traceable aut...   \n",
       "2        0000902 // cell morphogenesis // inferred from...   \n",
       "3        0001655 // urogenital system development // in...   \n",
       "4        0007165 // signal transduction // non-traceabl...   \n",
       "...                                                    ...   \n",
       "2679118                                                NaN   \n",
       "2679119                                                NaN   \n",
       "2679120                                                NaN   \n",
       "2679121                                                NaN   \n",
       "2679122                                                NaN   \n",
       "\n",
       "                          Gene Ontology Cellular Component  \\\n",
       "0        0005576 // extracellular region // inferred fr...   \n",
       "1        0005634 // nucleus // inferred from electronic...   \n",
       "2        0005737 // cytoplasm // inferred from direct a...   \n",
       "3        0005634 // nucleus // inferred from direct ass...   \n",
       "4        0001750 // photoreceptor outer segment // infe...   \n",
       "...                                                    ...   \n",
       "2679118                                                NaN   \n",
       "2679119                                                NaN   \n",
       "2679120                                                NaN   \n",
       "2679121                                                NaN   \n",
       "2679122                                                NaN   \n",
       "\n",
       "                          Gene Ontology Molecular Function  \n",
       "0        0000166 // nucleotide binding // inferred from...  \n",
       "1        0000166 // nucleotide binding // inferred from...  \n",
       "2        0000166 // nucleotide binding // inferred from...  \n",
       "3        0000979 // RNA polymerase II core promoter seq...  \n",
       "4        0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                    ...  \n",
       "2679118                                                NaN  \n",
       "2679119                                                NaN  \n",
       "2679120                                                NaN  \n",
       "2679121                                                NaN  \n",
       "2679122                                                NaN  \n",
       "\n",
       "[2679123 rows x 16 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>1341.22</td>\n",
       "      <td>1980.88</td>\n",
       "      <td>1241.750</td>\n",
       "      <td>1284.08</td>\n",
       "      <td>455.520</td>\n",
       "      <td>15.020</td>\n",
       "      <td>8.490</td>\n",
       "      <td>2015.63</td>\n",
       "      <td>369.640</td>\n",
       "      <td>56.960</td>\n",
       "      <td>...</td>\n",
       "      <td>333.370</td>\n",
       "      <td>2458.34</td>\n",
       "      <td>180.17</td>\n",
       "      <td>736.710</td>\n",
       "      <td>460.15</td>\n",
       "      <td>878.59</td>\n",
       "      <td>460.98</td>\n",
       "      <td>8.25</td>\n",
       "      <td>161.38</td>\n",
       "      <td>526.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>7.83</td>\n",
       "      <td>10.19</td>\n",
       "      <td>13.070</td>\n",
       "      <td>11.64</td>\n",
       "      <td>76.920</td>\n",
       "      <td>11.640</td>\n",
       "      <td>23.720</td>\n",
       "      <td>13.40</td>\n",
       "      <td>5.810</td>\n",
       "      <td>23.780</td>\n",
       "      <td>...</td>\n",
       "      <td>5.770</td>\n",
       "      <td>9.76</td>\n",
       "      <td>7.01</td>\n",
       "      <td>5.500</td>\n",
       "      <td>8.54</td>\n",
       "      <td>6.73</td>\n",
       "      <td>7.71</td>\n",
       "      <td>10.09</td>\n",
       "      <td>11.64</td>\n",
       "      <td>11.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>7.83</td>\n",
       "      <td>10.19</td>\n",
       "      <td>13.070</td>\n",
       "      <td>11.64</td>\n",
       "      <td>76.920</td>\n",
       "      <td>11.640</td>\n",
       "      <td>23.720</td>\n",
       "      <td>13.40</td>\n",
       "      <td>5.810</td>\n",
       "      <td>23.780</td>\n",
       "      <td>...</td>\n",
       "      <td>5.770</td>\n",
       "      <td>9.76</td>\n",
       "      <td>7.01</td>\n",
       "      <td>5.500</td>\n",
       "      <td>8.54</td>\n",
       "      <td>6.73</td>\n",
       "      <td>7.71</td>\n",
       "      <td>10.09</td>\n",
       "      <td>11.64</td>\n",
       "      <td>11.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1P2</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.83</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC078883.4</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>6.83</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abParts</th>\n",
       "      <td>9.04</td>\n",
       "      <td>9.04</td>\n",
       "      <td>9.040</td>\n",
       "      <td>9.04</td>\n",
       "      <td>35.270</td>\n",
       "      <td>6313.010</td>\n",
       "      <td>233.550</td>\n",
       "      <td>9.04</td>\n",
       "      <td>9.040</td>\n",
       "      <td>9.040</td>\n",
       "      <td>...</td>\n",
       "      <td>9.040</td>\n",
       "      <td>9.04</td>\n",
       "      <td>9.05</td>\n",
       "      <td>9.040</td>\n",
       "      <td>11.09</td>\n",
       "      <td>9.61</td>\n",
       "      <td>9.04</td>\n",
       "      <td>52.44</td>\n",
       "      <td>43.59</td>\n",
       "      <td>9.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>13.290</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>44.860</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>av27s1</th>\n",
       "      <td>5.50</td>\n",
       "      <td>5.72</td>\n",
       "      <td>5.640</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>9.31</td>\n",
       "      <td>9.05</td>\n",
       "      <td>8.065</td>\n",
       "      <td>6.64</td>\n",
       "      <td>8.985</td>\n",
       "      <td>11.525</td>\n",
       "      <td>11.875</td>\n",
       "      <td>8.55</td>\n",
       "      <td>20.685</td>\n",
       "      <td>6.775</td>\n",
       "      <td>...</td>\n",
       "      <td>5.625</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mir-223</th>\n",
       "      <td>26.66</td>\n",
       "      <td>7.34</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>6.250</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "      <td>5.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25851 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GSM493251  GSM493252  GSM493253  GSM493254  GSM493255  \\\n",
       "Gene                                                                   \n",
       " ABCB4           1341.22    1980.88   1241.750    1284.08    455.520   \n",
       " ABCC6P1            7.83      10.19     13.070      11.64     76.920   \n",
       " ABCC6P2            7.83      10.19     13.070      11.64     76.920   \n",
       " ABCD1P2            5.50       5.83      5.500       5.50      5.500   \n",
       " AC078883.4         5.50       5.50      5.500       5.50      5.500   \n",
       "...                  ...        ...        ...        ...        ...   \n",
       "abParts             9.04       9.04      9.040       9.04     35.270   \n",
       "alpha               5.50       5.50      5.500       5.50      5.500   \n",
       "av27s1              5.50       5.72      5.640       5.50      5.500   \n",
       "hsa-let-7a-3        9.31       9.05      8.065       6.64      8.985   \n",
       "mir-223            26.66       7.34      5.500       5.50      5.500   \n",
       "\n",
       "               GSM493256  GSM493257  GSM493258  GSM493259  GSM493260  ...  \\\n",
       "Gene                                                                  ...   \n",
       " ABCB4            15.020      8.490    2015.63    369.640     56.960  ...   \n",
       " ABCC6P1          11.640     23.720      13.40      5.810     23.780  ...   \n",
       " ABCC6P2          11.640     23.720      13.40      5.810     23.780  ...   \n",
       " ABCD1P2           5.500      5.500       5.50      5.500      5.500  ...   \n",
       " AC078883.4        5.500      5.500       5.50      5.500      5.500  ...   \n",
       "...                  ...        ...        ...        ...        ...  ...   \n",
       "abParts         6313.010    233.550       9.04      9.040      9.040  ...   \n",
       "alpha             13.290      5.500       5.50     44.860      5.500  ...   \n",
       "av27s1             5.500      5.500       5.50      5.500      5.500  ...   \n",
       "hsa-let-7a-3      11.525     11.875       8.55     20.685      6.775  ...   \n",
       "mir-223            6.250      5.500       5.50      5.500      5.500  ...   \n",
       "\n",
       "               GSM1094071  GSM1094072  GSM1094073  GSM1094074  GSM1094075  \\\n",
       "Gene                                                                        \n",
       " ABCB4            333.370     2458.34      180.17     736.710      460.15   \n",
       " ABCC6P1            5.770        9.76        7.01       5.500        8.54   \n",
       " ABCC6P2            5.770        9.76        7.01       5.500        8.54   \n",
       " ABCD1P2            5.500        5.50        5.50       5.500        5.50   \n",
       " AC078883.4         5.500        5.50        5.50       5.500        5.50   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts             9.040        9.04        9.05       9.040       11.09   \n",
       "alpha               5.500        5.50        5.50       5.500        5.50   \n",
       "av27s1              5.500        5.50        5.50       5.500        5.50   \n",
       "hsa-let-7a-3        5.625        5.50        5.50       5.675        5.50   \n",
       "mir-223             5.500        5.50        5.50       5.500        5.50   \n",
       "\n",
       "               GSM1094076  GSM1094077  GSM1094078  GSM1094079  GSM1094080  \n",
       "Gene                                                                       \n",
       " ABCB4             878.59      460.98        8.25      161.38      526.21  \n",
       " ABCC6P1             6.73        7.71       10.09       11.64       11.64  \n",
       " ABCC6P2             6.73        7.71       10.09       11.64       11.64  \n",
       " ABCD1P2             5.50        5.50        5.50        5.50        5.50  \n",
       " AC078883.4          6.83        5.50        5.50        5.50        5.50  \n",
       "...                   ...         ...         ...         ...         ...  \n",
       "abParts              9.61        9.04       52.44       43.59        9.04  \n",
       "alpha                5.50        5.50        5.50        5.50        5.50  \n",
       "av27s1               5.50        5.50        5.50        5.50        5.50  \n",
       "hsa-let-7a-3         5.50        5.50        5.50        5.50        5.50  \n",
       "mir-223              5.50        5.50        5.50        5.50        5.50  \n",
       "\n",
       "[25851 rows x 48 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM493251</th>\n",
       "      <th>GSM493252</th>\n",
       "      <th>GSM493253</th>\n",
       "      <th>GSM493254</th>\n",
       "      <th>GSM493255</th>\n",
       "      <th>GSM493256</th>\n",
       "      <th>GSM493257</th>\n",
       "      <th>GSM493258</th>\n",
       "      <th>GSM493259</th>\n",
       "      <th>GSM493260</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1094071</th>\n",
       "      <th>GSM1094072</th>\n",
       "      <th>GSM1094073</th>\n",
       "      <th>GSM1094074</th>\n",
       "      <th>GSM1094075</th>\n",
       "      <th>GSM1094076</th>\n",
       "      <th>GSM1094077</th>\n",
       "      <th>GSM1094078</th>\n",
       "      <th>GSM1094079</th>\n",
       "      <th>GSM1094080</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.640</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>18.570000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>6.410000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.980000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>8.410000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>7.950000</td>\n",
       "      <td>12.980</td>\n",
       "      <td>13.340000</td>\n",
       "      <td>8.350000</td>\n",
       "      <td>8.840000</td>\n",
       "      <td>7.170000</td>\n",
       "      <td>10.140000</td>\n",
       "      <td>6.800000</td>\n",
       "      <td>6.540000</td>\n",
       "      <td>15.120000</td>\n",
       "      <td>...</td>\n",
       "      <td>10.540000</td>\n",
       "      <td>8.270000</td>\n",
       "      <td>16.220000</td>\n",
       "      <td>7.580000</td>\n",
       "      <td>9.610000</td>\n",
       "      <td>13.200000</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>39.740000</td>\n",
       "      <td>7.510000</td>\n",
       "      <td>6.300000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>7.120000</td>\n",
       "      <td>5.500</td>\n",
       "      <td>6.805000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.930000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>6.555000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>3303.540000</td>\n",
       "      <td>1751.345</td>\n",
       "      <td>4304.375000</td>\n",
       "      <td>4467.760000</td>\n",
       "      <td>1225.825000</td>\n",
       "      <td>404.895000</td>\n",
       "      <td>276.810000</td>\n",
       "      <td>922.500000</td>\n",
       "      <td>184.890000</td>\n",
       "      <td>296.650000</td>\n",
       "      <td>...</td>\n",
       "      <td>503.735000</td>\n",
       "      <td>832.570000</td>\n",
       "      <td>1589.435000</td>\n",
       "      <td>924.025000</td>\n",
       "      <td>467.175000</td>\n",
       "      <td>984.970000</td>\n",
       "      <td>678.560000</td>\n",
       "      <td>249.025000</td>\n",
       "      <td>1012.500000</td>\n",
       "      <td>963.050000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>13.250000</td>\n",
       "      <td>10.690</td>\n",
       "      <td>33.790000</td>\n",
       "      <td>23.860000</td>\n",
       "      <td>27.390000</td>\n",
       "      <td>9.860000</td>\n",
       "      <td>8.260000</td>\n",
       "      <td>49.400000</td>\n",
       "      <td>9.820000</td>\n",
       "      <td>14.360000</td>\n",
       "      <td>...</td>\n",
       "      <td>17.580000</td>\n",
       "      <td>24.270000</td>\n",
       "      <td>27.740000</td>\n",
       "      <td>16.180000</td>\n",
       "      <td>8.800000</td>\n",
       "      <td>16.650000</td>\n",
       "      <td>19.140000</td>\n",
       "      <td>9.390000</td>\n",
       "      <td>21.520000</td>\n",
       "      <td>18.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.680000</td>\n",
       "      <td>41.110000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>121.390000</td>\n",
       "      <td>5.500000</td>\n",
       "      <td>6.110000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>120.173333</td>\n",
       "      <td>135.650</td>\n",
       "      <td>144.273333</td>\n",
       "      <td>160.646667</td>\n",
       "      <td>192.403333</td>\n",
       "      <td>69.213333</td>\n",
       "      <td>164.600000</td>\n",
       "      <td>153.056667</td>\n",
       "      <td>118.900000</td>\n",
       "      <td>124.290000</td>\n",
       "      <td>...</td>\n",
       "      <td>52.633333</td>\n",
       "      <td>102.250000</td>\n",
       "      <td>73.350000</td>\n",
       "      <td>83.226667</td>\n",
       "      <td>84.833333</td>\n",
       "      <td>115.490000</td>\n",
       "      <td>64.146667</td>\n",
       "      <td>122.486667</td>\n",
       "      <td>66.980000</td>\n",
       "      <td>111.153333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>109.345000</td>\n",
       "      <td>103.390</td>\n",
       "      <td>115.915000</td>\n",
       "      <td>113.375000</td>\n",
       "      <td>33.390000</td>\n",
       "      <td>169.690000</td>\n",
       "      <td>97.565000</td>\n",
       "      <td>68.930000</td>\n",
       "      <td>105.195000</td>\n",
       "      <td>18.495000</td>\n",
       "      <td>...</td>\n",
       "      <td>18.360000</td>\n",
       "      <td>28.040000</td>\n",
       "      <td>86.410000</td>\n",
       "      <td>41.670000</td>\n",
       "      <td>31.345000</td>\n",
       "      <td>24.975000</td>\n",
       "      <td>43.145000</td>\n",
       "      <td>16.535000</td>\n",
       "      <td>16.925000</td>\n",
       "      <td>19.220000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>46.093333</td>\n",
       "      <td>37.000</td>\n",
       "      <td>36.803333</td>\n",
       "      <td>34.036667</td>\n",
       "      <td>53.216667</td>\n",
       "      <td>21.713333</td>\n",
       "      <td>37.506667</td>\n",
       "      <td>89.816667</td>\n",
       "      <td>77.906667</td>\n",
       "      <td>36.686667</td>\n",
       "      <td>...</td>\n",
       "      <td>58.003333</td>\n",
       "      <td>15.563333</td>\n",
       "      <td>8.223333</td>\n",
       "      <td>11.126667</td>\n",
       "      <td>9.826667</td>\n",
       "      <td>10.496667</td>\n",
       "      <td>13.913333</td>\n",
       "      <td>11.370000</td>\n",
       "      <td>11.766667</td>\n",
       "      <td>15.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>54.605000</td>\n",
       "      <td>61.375</td>\n",
       "      <td>60.770000</td>\n",
       "      <td>61.425000</td>\n",
       "      <td>89.515000</td>\n",
       "      <td>119.045000</td>\n",
       "      <td>242.765000</td>\n",
       "      <td>115.675000</td>\n",
       "      <td>71.390000</td>\n",
       "      <td>100.650000</td>\n",
       "      <td>...</td>\n",
       "      <td>80.365000</td>\n",
       "      <td>107.690000</td>\n",
       "      <td>74.930000</td>\n",
       "      <td>66.605000</td>\n",
       "      <td>115.315000</td>\n",
       "      <td>64.875000</td>\n",
       "      <td>54.920000</td>\n",
       "      <td>206.205000</td>\n",
       "      <td>62.815000</td>\n",
       "      <td>64.870000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM493251  GSM493252    GSM493253    GSM493254    GSM493255  \\\n",
       "A1BG         5.500000      5.640     5.500000     5.500000     5.500000   \n",
       "A1BG-AS1     7.950000     12.980    13.340000     8.350000     8.840000   \n",
       "A1CF         7.120000      5.500     6.805000     5.500000     5.500000   \n",
       "A2M       3303.540000   1751.345  4304.375000  4467.760000  1225.825000   \n",
       "A2M-AS1     13.250000     10.690    33.790000    23.860000    27.390000   \n",
       "...               ...        ...          ...          ...          ...   \n",
       "ZYG11A       5.500000      5.500     5.500000     5.500000     5.500000   \n",
       "ZYG11B     120.173333    135.650   144.273333   160.646667   192.403333   \n",
       "ZYX        109.345000    103.390   115.915000   113.375000    33.390000   \n",
       "ZZEF1       46.093333     37.000    36.803333    34.036667    53.216667   \n",
       "ZZZ3        54.605000     61.375    60.770000    61.425000    89.515000   \n",
       "\n",
       "           GSM493256   GSM493257   GSM493258   GSM493259   GSM493260  ...  \\\n",
       "A1BG        5.500000   18.570000    5.500000    5.500000    5.500000  ...   \n",
       "A1BG-AS1    7.170000   10.140000    6.800000    6.540000   15.120000  ...   \n",
       "A1CF        5.500000    5.930000    5.500000    6.555000    5.500000  ...   \n",
       "A2M       404.895000  276.810000  922.500000  184.890000  296.650000  ...   \n",
       "A2M-AS1     9.860000    8.260000   49.400000    9.820000   14.360000  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A      5.680000   41.110000    5.500000    5.500000    5.500000  ...   \n",
       "ZYG11B     69.213333  164.600000  153.056667  118.900000  124.290000  ...   \n",
       "ZYX       169.690000   97.565000   68.930000  105.195000   18.495000  ...   \n",
       "ZZEF1      21.713333   37.506667   89.816667   77.906667   36.686667  ...   \n",
       "ZZZ3      119.045000  242.765000  115.675000   71.390000  100.650000  ...   \n",
       "\n",
       "          GSM1094071  GSM1094072   GSM1094073  GSM1094074  GSM1094075  \\\n",
       "A1BG        5.500000    5.500000     6.410000    5.500000    5.500000   \n",
       "A1BG-AS1   10.540000    8.270000    16.220000    7.580000    9.610000   \n",
       "A1CF        5.500000    5.500000     5.500000    5.500000    5.500000   \n",
       "A2M       503.735000  832.570000  1589.435000  924.025000  467.175000   \n",
       "A2M-AS1    17.580000   24.270000    27.740000   16.180000    8.800000   \n",
       "...              ...         ...          ...         ...         ...   \n",
       "ZYG11A      5.500000    5.500000     5.500000    5.500000    5.500000   \n",
       "ZYG11B     52.633333  102.250000    73.350000   83.226667   84.833333   \n",
       "ZYX        18.360000   28.040000    86.410000   41.670000   31.345000   \n",
       "ZZEF1      58.003333   15.563333     8.223333   11.126667    9.826667   \n",
       "ZZZ3       80.365000  107.690000    74.930000   66.605000  115.315000   \n",
       "\n",
       "          GSM1094076  GSM1094077  GSM1094078   GSM1094079  GSM1094080  \n",
       "A1BG        5.980000    5.500000    8.410000     5.500000    5.500000  \n",
       "A1BG-AS1   13.200000    6.500000   39.740000     7.510000    6.300000  \n",
       "A1CF        5.500000    5.500000    5.500000     5.500000    5.500000  \n",
       "A2M       984.970000  678.560000  249.025000  1012.500000  963.050000  \n",
       "A2M-AS1    16.650000   19.140000    9.390000    21.520000   18.600000  \n",
       "...              ...         ...         ...          ...         ...  \n",
       "ZYG11A      5.500000    5.500000  121.390000     5.500000    6.110000  \n",
       "ZYG11B    115.490000   64.146667  122.486667    66.980000  111.153333  \n",
       "ZYX        24.975000   43.145000   16.535000    16.925000   19.220000  \n",
       "ZZEF1      10.496667   13.913333   11.370000    11.766667   15.500000  \n",
       "ZZZ3       64.875000   54.920000  206.205000    62.815000   64.870000  \n",
       "\n",
       "[17857 rows x 48 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:718: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n"
     ]
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM493265</th>\n",
       "      <td>1.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.53</td>\n",
       "      <td>27.98</td>\n",
       "      <td>5.5</td>\n",
       "      <td>630.915</td>\n",
       "      <td>14.83</td>\n",
       "      <td>5.84</td>\n",
       "      <td>5.79</td>\n",
       "      <td>...</td>\n",
       "      <td>21.53</td>\n",
       "      <td>86.84</td>\n",
       "      <td>8.83</td>\n",
       "      <td>6.735</td>\n",
       "      <td>36.676667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>98.686667</td>\n",
       "      <td>34.93</td>\n",
       "      <td>30.373333</td>\n",
       "      <td>71.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM493270</th>\n",
       "      <td>1.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.77</td>\n",
       "      <td>5.5</td>\n",
       "      <td>968.32</td>\n",
       "      <td>34.95</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5.79</td>\n",
       "      <td>...</td>\n",
       "      <td>30.695</td>\n",
       "      <td>90.66</td>\n",
       "      <td>13.94</td>\n",
       "      <td>11.26</td>\n",
       "      <td>33.223333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>158.316667</td>\n",
       "      <td>70.065</td>\n",
       "      <td>61.263333</td>\n",
       "      <td>68.105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094056</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>9.68</td>\n",
       "      <td>5.5</td>\n",
       "      <td>380.405</td>\n",
       "      <td>9.32</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5.84</td>\n",
       "      <td>...</td>\n",
       "      <td>48.445</td>\n",
       "      <td>191.94</td>\n",
       "      <td>12.93</td>\n",
       "      <td>9.78</td>\n",
       "      <td>41.331667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>89.596667</td>\n",
       "      <td>15.265</td>\n",
       "      <td>16.493333</td>\n",
       "      <td>108.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094057</th>\n",
       "      <td>1.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.95</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1465.11</td>\n",
       "      <td>27.9</td>\n",
       "      <td>5.5</td>\n",
       "      <td>154.36</td>\n",
       "      <td>...</td>\n",
       "      <td>48.365</td>\n",
       "      <td>281.2</td>\n",
       "      <td>7.29</td>\n",
       "      <td>7.32</td>\n",
       "      <td>46.606667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>50.603333</td>\n",
       "      <td>28.48</td>\n",
       "      <td>10.89</td>\n",
       "      <td>58.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094060</th>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.95</td>\n",
       "      <td>5.5</td>\n",
       "      <td>680.91</td>\n",
       "      <td>14.43</td>\n",
       "      <td>5.5</td>\n",
       "      <td>12.75</td>\n",
       "      <td>...</td>\n",
       "      <td>47.235</td>\n",
       "      <td>28.79</td>\n",
       "      <td>10.14</td>\n",
       "      <td>11.98</td>\n",
       "      <td>45.533333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>70.15</td>\n",
       "      <td>15.255</td>\n",
       "      <td>23.983333</td>\n",
       "      <td>101.235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094061</th>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.29</td>\n",
       "      <td>5.5</td>\n",
       "      <td>999.32</td>\n",
       "      <td>11.21</td>\n",
       "      <td>5.5</td>\n",
       "      <td>18.01</td>\n",
       "      <td>...</td>\n",
       "      <td>32.09</td>\n",
       "      <td>81.36</td>\n",
       "      <td>19.22</td>\n",
       "      <td>9.825</td>\n",
       "      <td>28.378333</td>\n",
       "      <td>5.56</td>\n",
       "      <td>76.246667</td>\n",
       "      <td>54.955</td>\n",
       "      <td>10.566667</td>\n",
       "      <td>67.585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094063</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.51</td>\n",
       "      <td>5.5</td>\n",
       "      <td>474.915</td>\n",
       "      <td>12.02</td>\n",
       "      <td>5.5</td>\n",
       "      <td>46.27</td>\n",
       "      <td>...</td>\n",
       "      <td>41.44</td>\n",
       "      <td>61.6</td>\n",
       "      <td>23.76</td>\n",
       "      <td>18.295</td>\n",
       "      <td>43.958333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>87.03</td>\n",
       "      <td>23.975</td>\n",
       "      <td>10.793333</td>\n",
       "      <td>94.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094066</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.71</td>\n",
       "      <td>22.0</td>\n",
       "      <td>5.945</td>\n",
       "      <td>365.945</td>\n",
       "      <td>11.49</td>\n",
       "      <td>5.56</td>\n",
       "      <td>60.55</td>\n",
       "      <td>...</td>\n",
       "      <td>35.47</td>\n",
       "      <td>65.86</td>\n",
       "      <td>32.64</td>\n",
       "      <td>25.8</td>\n",
       "      <td>55.596667</td>\n",
       "      <td>7.74</td>\n",
       "      <td>89.68</td>\n",
       "      <td>12.18</td>\n",
       "      <td>14.486667</td>\n",
       "      <td>95.235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094067</th>\n",
       "      <td>0.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.2</td>\n",
       "      <td>6.55</td>\n",
       "      <td>1191.28</td>\n",
       "      <td>23.69</td>\n",
       "      <td>5.5</td>\n",
       "      <td>9.22</td>\n",
       "      <td>...</td>\n",
       "      <td>33.985</td>\n",
       "      <td>37.59</td>\n",
       "      <td>24.58</td>\n",
       "      <td>22.49</td>\n",
       "      <td>27.738333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>109.716667</td>\n",
       "      <td>42.44</td>\n",
       "      <td>10.15</td>\n",
       "      <td>61.765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094071</th>\n",
       "      <td>0.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>10.54</td>\n",
       "      <td>5.5</td>\n",
       "      <td>503.735</td>\n",
       "      <td>17.58</td>\n",
       "      <td>5.5</td>\n",
       "      <td>93.73</td>\n",
       "      <td>...</td>\n",
       "      <td>53.105</td>\n",
       "      <td>159.57</td>\n",
       "      <td>14.95</td>\n",
       "      <td>11.98</td>\n",
       "      <td>41.068333</td>\n",
       "      <td>5.5</td>\n",
       "      <td>52.633333</td>\n",
       "      <td>18.36</td>\n",
       "      <td>58.003333</td>\n",
       "      <td>80.365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094074</th>\n",
       "      <td>1.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>7.58</td>\n",
       "      <td>5.5</td>\n",
       "      <td>924.025</td>\n",
       "      <td>16.18</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.8</td>\n",
       "      <td>...</td>\n",
       "      <td>64.32</td>\n",
       "      <td>139.03</td>\n",
       "      <td>20.97</td>\n",
       "      <td>19.525</td>\n",
       "      <td>33.635</td>\n",
       "      <td>5.5</td>\n",
       "      <td>83.226667</td>\n",
       "      <td>41.67</td>\n",
       "      <td>11.126667</td>\n",
       "      <td>66.605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094075</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.5</td>\n",
       "      <td>9.61</td>\n",
       "      <td>5.5</td>\n",
       "      <td>467.175</td>\n",
       "      <td>8.8</td>\n",
       "      <td>5.5</td>\n",
       "      <td>6.38</td>\n",
       "      <td>...</td>\n",
       "      <td>38.765</td>\n",
       "      <td>92.65</td>\n",
       "      <td>49.51</td>\n",
       "      <td>48.36</td>\n",
       "      <td>31.856667</td>\n",
       "      <td>5.5</td>\n",
       "      <td>84.833333</td>\n",
       "      <td>31.345</td>\n",
       "      <td>9.826667</td>\n",
       "      <td>115.315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1094078</th>\n",
       "      <td>1.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.41</td>\n",
       "      <td>39.74</td>\n",
       "      <td>5.5</td>\n",
       "      <td>249.025</td>\n",
       "      <td>9.39</td>\n",
       "      <td>5.5</td>\n",
       "      <td>5.97</td>\n",
       "      <td>...</td>\n",
       "      <td>179.07</td>\n",
       "      <td>240.64</td>\n",
       "      <td>17.37</td>\n",
       "      <td>22.465</td>\n",
       "      <td>46.6</td>\n",
       "      <td>121.39</td>\n",
       "      <td>122.486667</td>\n",
       "      <td>16.535</td>\n",
       "      <td>11.37</td>\n",
       "      <td>206.205</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows Ã— 17860 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer   Age Gender   A1BG A1BG-AS1   A1CF      A2M  \\\n",
       "GSM493265                    1.0  37.0    1.0  29.53    27.98    5.5  630.915   \n",
       "GSM493270                    1.0  58.0    1.0    5.5     6.77    5.5   968.32   \n",
       "GSM1094056                   0.0  20.0    1.0    5.5     9.68    5.5  380.405   \n",
       "GSM1094057                   1.0  68.0    1.0    5.5     7.95    5.5  1465.11   \n",
       "GSM1094060                   0.0  32.0    1.0    5.5     7.95    5.5   680.91   \n",
       "GSM1094061                   0.0  43.0    0.0    5.5     6.29    5.5   999.32   \n",
       "GSM1094063                   1.0  40.0    0.0    5.5     7.51    5.5  474.915   \n",
       "GSM1094066                   0.0  27.0    1.0  10.71     22.0  5.945  365.945   \n",
       "GSM1094067                   0.0  70.0    0.0    5.5      7.2   6.55  1191.28   \n",
       "GSM1094071                   0.0  57.0    0.0    5.5    10.54    5.5  503.735   \n",
       "GSM1094074                   1.0  55.0    1.0    5.5     7.58    5.5  924.025   \n",
       "GSM1094075                   1.0  51.0    1.0    5.5     9.61    5.5  467.175   \n",
       "GSM1094078                   1.0  63.0    0.0   8.41    39.74    5.5  249.025   \n",
       "\n",
       "           A2M-AS1 A2ML1   A2MP1  ...  ZWILCH   ZWINT   ZXDA    ZXDB  \\\n",
       "GSM493265    14.83  5.84    5.79  ...   21.53   86.84   8.83   6.735   \n",
       "GSM493270    34.95   5.5    5.79  ...  30.695   90.66  13.94   11.26   \n",
       "GSM1094056    9.32   5.5    5.84  ...  48.445  191.94  12.93    9.78   \n",
       "GSM1094057    27.9   5.5  154.36  ...  48.365   281.2   7.29    7.32   \n",
       "GSM1094060   14.43   5.5   12.75  ...  47.235   28.79  10.14   11.98   \n",
       "GSM1094061   11.21   5.5   18.01  ...   32.09   81.36  19.22   9.825   \n",
       "GSM1094063   12.02   5.5   46.27  ...   41.44    61.6  23.76  18.295   \n",
       "GSM1094066   11.49  5.56   60.55  ...   35.47   65.86  32.64    25.8   \n",
       "GSM1094067   23.69   5.5    9.22  ...  33.985   37.59  24.58   22.49   \n",
       "GSM1094071   17.58   5.5   93.73  ...  53.105  159.57  14.95   11.98   \n",
       "GSM1094074   16.18   5.5     6.8  ...   64.32  139.03  20.97  19.525   \n",
       "GSM1094075     8.8   5.5    6.38  ...  38.765   92.65  49.51   48.36   \n",
       "GSM1094078    9.39   5.5    5.97  ...  179.07  240.64  17.37  22.465   \n",
       "\n",
       "                 ZXDC  ZYG11A      ZYG11B     ZYX      ZZEF1     ZZZ3  \n",
       "GSM493265   36.676667     5.5   98.686667   34.93  30.373333    71.49  \n",
       "GSM493270   33.223333     5.5  158.316667  70.065  61.263333   68.105  \n",
       "GSM1094056  41.331667     5.5   89.596667  15.265  16.493333   108.55  \n",
       "GSM1094057  46.606667     5.5   50.603333   28.48      10.89    58.92  \n",
       "GSM1094060  45.533333     5.5       70.15  15.255  23.983333  101.235  \n",
       "GSM1094061  28.378333    5.56   76.246667  54.955  10.566667   67.585  \n",
       "GSM1094063  43.958333     5.5       87.03  23.975  10.793333    94.43  \n",
       "GSM1094066  55.596667    7.74       89.68   12.18  14.486667   95.235  \n",
       "GSM1094067  27.738333     5.5  109.716667   42.44      10.15   61.765  \n",
       "GSM1094071  41.068333     5.5   52.633333   18.36  58.003333   80.365  \n",
       "GSM1094074     33.635     5.5   83.226667   41.67  11.126667   66.605  \n",
       "GSM1094075  31.856667     5.5   84.833333  31.345   9.826667  115.315  \n",
       "GSM1094078       46.6  121.39  122.486667  16.535      11.37  206.205  \n",
       "\n",
       "[13 rows x 17860 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 13 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '0.0' with 6 occurrences. This represents 46.15% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 37.0\n",
      "  50% (Median): 51.0\n",
      "  75%: 58.0\n",
      "Min: 20.0\n",
      "Max: 70.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 5 occurrences. This represents 38.46% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Comprehensive molecular profiling of children with recurrent cancer\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2889381</th>\n",
       "      <th>GSM2889382</th>\n",
       "      <th>GSM2889383</th>\n",
       "      <th>GSM2889384</th>\n",
       "      <th>GSM2889385</th>\n",
       "      <th>GSM2889386</th>\n",
       "      <th>GSM2889387</th>\n",
       "      <th>GSM2889388</th>\n",
       "      <th>GSM2889389</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2889414</th>\n",
       "      <th>GSM2889415</th>\n",
       "      <th>GSM2889416</th>\n",
       "      <th>GSM2889417</th>\n",
       "      <th>GSM2889418</th>\n",
       "      <th>GSM2889419</th>\n",
       "      <th>GSM2889420</th>\n",
       "      <th>GSM2889421</th>\n",
       "      <th>GSM2889422</th>\n",
       "      <th>GSM2889423</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>condition: Atypical meningioma</td>\n",
       "      <td>condition: Choroid plexus carcinoma / Malignan...</td>\n",
       "      <td>condition: Pilocytisc/pilomyxoid astrocytoma</td>\n",
       "      <td>condition: Pleomorphic xanthoastrocytoma</td>\n",
       "      <td>condition: Mesoblastisc nephroma</td>\n",
       "      <td>condition: Signetringcell carcinoma</td>\n",
       "      <td>condition: Ganglioglioma / Diffuse astrocytoma</td>\n",
       "      <td>condition: Chondrosarkoma</td>\n",
       "      <td>condition: Chordoma, dedefferentiated/anaplati...</td>\n",
       "      <td>...</td>\n",
       "      <td>condition: Anaplastic ependymoma</td>\n",
       "      <td>condition: Enchodroms</td>\n",
       "      <td>condition: Pineoblastoma</td>\n",
       "      <td>condition: Osteochondroma</td>\n",
       "      <td>condition: Malignant peripheral nerve sheeth t...</td>\n",
       "      <td>condition: Ewing sarcoma</td>\n",
       "      <td>condition: Adrenocortical carcinoma</td>\n",
       "      <td>condition: Anaplastic ependymoma</td>\n",
       "      <td>condition: Rhabdomyosarcoma</td>\n",
       "      <td>condition: Ependymom</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                      GSM2889381  \\\n",
       "0  !Sample_characteristics_ch1  condition: Atypical meningioma   \n",
       "\n",
       "                                          GSM2889382  \\\n",
       "0  condition: Choroid plexus carcinoma / Malignan...   \n",
       "\n",
       "                                     GSM2889383  \\\n",
       "0  condition: Pilocytisc/pilomyxoid astrocytoma   \n",
       "\n",
       "                                 GSM2889384                        GSM2889385  \\\n",
       "0  condition: Pleomorphic xanthoastrocytoma  condition: Mesoblastisc nephroma   \n",
       "\n",
       "                            GSM2889386  \\\n",
       "0  condition: Signetringcell carcinoma   \n",
       "\n",
       "                                       GSM2889387                 GSM2889388  \\\n",
       "0  condition: Ganglioglioma / Diffuse astrocytoma  condition: Chondrosarkoma   \n",
       "\n",
       "                                          GSM2889389  ...  \\\n",
       "0  condition: Chordoma, dedefferentiated/anaplati...  ...   \n",
       "\n",
       "                         GSM2889414             GSM2889415  \\\n",
       "0  condition: Anaplastic ependymoma  condition: Enchodroms   \n",
       "\n",
       "                 GSM2889416                 GSM2889417  \\\n",
       "0  condition: Pineoblastoma  condition: Osteochondroma   \n",
       "\n",
       "                                          GSM2889418  \\\n",
       "0  condition: Malignant peripheral nerve sheeth t...   \n",
       "\n",
       "                 GSM2889419                           GSM2889420  \\\n",
       "0  condition: Ewing sarcoma  condition: Adrenocortical carcinoma   \n",
       "\n",
       "                         GSM2889421                   GSM2889422  \\\n",
       "0  condition: Anaplastic ependymoma  condition: Rhabdomyosarcoma   \n",
       "\n",
       "             GSM2889423  \n",
       "0  condition: Ependymom  \n",
       "\n",
       "[1 rows x 44 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stoped: MemoryError\n",
    "\n",
    "cohort = accession_num = \"GSE108089\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Genome-scale methylome profiling of adrenocortical carcinomas (ACC) and adenomas (ACA)\"\n",
      "!Series_summary\t\"Genome-scale DNA methylation was analyzed in a cohort of ACC and ACA to identify DNA methylation changes.\"\n",
      "!Series_overall_design\t\"Bisulfite converted DNA from 51 fresh frozen ACC and 30 ACA samples were hybridized to Illumina HumanMethylation27 BeadChips.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1196428</th>\n",
       "      <th>GSM1196429</th>\n",
       "      <th>GSM1196430</th>\n",
       "      <th>GSM1196431</th>\n",
       "      <th>GSM1196432</th>\n",
       "      <th>GSM1196433</th>\n",
       "      <th>GSM1196434</th>\n",
       "      <th>GSM1196435</th>\n",
       "      <th>GSM1196436</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196501</th>\n",
       "      <th>GSM1196502</th>\n",
       "      <th>GSM1196503</th>\n",
       "      <th>GSM1196504</th>\n",
       "      <th>GSM1196505</th>\n",
       "      <th>GSM1196506</th>\n",
       "      <th>GSM1196507</th>\n",
       "      <th>GSM1196508</th>\n",
       "      <th>GSM1196509</th>\n",
       "      <th>GSM1196510</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "      <td>cell type: Adrenocortical adenoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM1196428  \\\n",
       "0  !Sample_characteristics_ch1  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196429                           GSM1196430  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196431                           GSM1196432  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196433                           GSM1196434  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196435                           GSM1196436  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "   ...                         GSM1196501                         GSM1196502  \\\n",
       "0  ...  cell type: Adrenocortical adenoma  cell type: Adrenocortical adenoma   \n",
       "\n",
       "                          GSM1196503                         GSM1196504  \\\n",
       "0  cell type: Adrenocortical adenoma  cell type: Adrenocortical adenoma   \n",
       "\n",
       "                          GSM1196505                         GSM1196506  \\\n",
       "0  cell type: Adrenocortical adenoma  cell type: Adrenocortical adenoma   \n",
       "\n",
       "                          GSM1196507                         GSM1196508  \\\n",
       "0  cell type: Adrenocortical adenoma  cell type: Adrenocortical adenoma   \n",
       "\n",
       "                          GSM1196509                         GSM1196510  \n",
       "0  cell type: Adrenocortical adenoma  cell type: Adrenocortical adenoma  \n",
       "\n",
       "[1 rows x 82 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "\n",
    "cohort = accession_num = \"GSE49277\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'cell type: Adrenocortical carcinoma',\n",
       "       'cell type: Adrenocortical adenoma'], dtype=object)"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'condition: Adrenocortical carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196428</th>\n",
       "      <th>GSM1196429</th>\n",
       "      <th>GSM1196430</th>\n",
       "      <th>GSM1196431</th>\n",
       "      <th>GSM1196432</th>\n",
       "      <th>GSM1196433</th>\n",
       "      <th>GSM1196434</th>\n",
       "      <th>GSM1196435</th>\n",
       "      <th>GSM1196436</th>\n",
       "      <th>GSM1196437</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196501</th>\n",
       "      <th>GSM1196502</th>\n",
       "      <th>GSM1196503</th>\n",
       "      <th>GSM1196504</th>\n",
       "      <th>GSM1196505</th>\n",
       "      <th>GSM1196506</th>\n",
       "      <th>GSM1196507</th>\n",
       "      <th>GSM1196508</th>\n",
       "      <th>GSM1196509</th>\n",
       "      <th>GSM1196510</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1196428  GSM1196429  GSM1196430  GSM1196431  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM1196432  GSM1196433  GSM1196434  GSM1196435  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM1196436  GSM1196437  ...  GSM1196501  GSM1196502  \\\n",
       "Adrenocortical Cancer           0           0  ...           0           0   \n",
       "\n",
       "                       GSM1196503  GSM1196504  GSM1196505  GSM1196506  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM1196507  GSM1196508  GSM1196509  GSM1196510  \n",
       "Adrenocortical Cancer           0           0           0           0  \n",
       "\n",
       "[1 rows x 81 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196428</th>\n",
       "      <th>GSM1196429</th>\n",
       "      <th>GSM1196430</th>\n",
       "      <th>GSM1196431</th>\n",
       "      <th>GSM1196432</th>\n",
       "      <th>GSM1196433</th>\n",
       "      <th>GSM1196434</th>\n",
       "      <th>GSM1196435</th>\n",
       "      <th>GSM1196436</th>\n",
       "      <th>GSM1196437</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196501</th>\n",
       "      <th>GSM1196502</th>\n",
       "      <th>GSM1196503</th>\n",
       "      <th>GSM1196504</th>\n",
       "      <th>GSM1196505</th>\n",
       "      <th>GSM1196506</th>\n",
       "      <th>GSM1196507</th>\n",
       "      <th>GSM1196508</th>\n",
       "      <th>GSM1196509</th>\n",
       "      <th>GSM1196510</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000292</th>\n",
       "      <td>0.853864</td>\n",
       "      <td>0.873688</td>\n",
       "      <td>0.835080</td>\n",
       "      <td>0.598675</td>\n",
       "      <td>0.770182</td>\n",
       "      <td>0.435826</td>\n",
       "      <td>0.779509</td>\n",
       "      <td>0.832013</td>\n",
       "      <td>0.787016</td>\n",
       "      <td>0.727145</td>\n",
       "      <td>...</td>\n",
       "      <td>0.584722</td>\n",
       "      <td>0.407948</td>\n",
       "      <td>0.590603</td>\n",
       "      <td>0.567782</td>\n",
       "      <td>0.359869</td>\n",
       "      <td>0.653657</td>\n",
       "      <td>0.771832</td>\n",
       "      <td>0.349547</td>\n",
       "      <td>0.693256</td>\n",
       "      <td>0.438942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00002426</th>\n",
       "      <td>0.183064</td>\n",
       "      <td>0.173539</td>\n",
       "      <td>0.205998</td>\n",
       "      <td>0.209957</td>\n",
       "      <td>0.176864</td>\n",
       "      <td>0.227453</td>\n",
       "      <td>0.194177</td>\n",
       "      <td>0.214679</td>\n",
       "      <td>0.097024</td>\n",
       "      <td>0.366682</td>\n",
       "      <td>...</td>\n",
       "      <td>0.217829</td>\n",
       "      <td>0.206476</td>\n",
       "      <td>0.169502</td>\n",
       "      <td>0.150175</td>\n",
       "      <td>0.228998</td>\n",
       "      <td>0.277141</td>\n",
       "      <td>0.248073</td>\n",
       "      <td>0.319916</td>\n",
       "      <td>0.258845</td>\n",
       "      <td>0.210580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00003994</th>\n",
       "      <td>0.146456</td>\n",
       "      <td>0.123106</td>\n",
       "      <td>0.045628</td>\n",
       "      <td>0.081798</td>\n",
       "      <td>0.190403</td>\n",
       "      <td>0.100151</td>\n",
       "      <td>0.060844</td>\n",
       "      <td>0.056040</td>\n",
       "      <td>0.060491</td>\n",
       "      <td>0.219284</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117100</td>\n",
       "      <td>0.054069</td>\n",
       "      <td>0.075588</td>\n",
       "      <td>0.076611</td>\n",
       "      <td>0.062579</td>\n",
       "      <td>0.188656</td>\n",
       "      <td>0.110269</td>\n",
       "      <td>0.058055</td>\n",
       "      <td>0.099765</td>\n",
       "      <td>0.113560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00005847</th>\n",
       "      <td>0.855151</td>\n",
       "      <td>0.811072</td>\n",
       "      <td>0.830661</td>\n",
       "      <td>0.512991</td>\n",
       "      <td>0.738095</td>\n",
       "      <td>0.367623</td>\n",
       "      <td>0.814609</td>\n",
       "      <td>0.681481</td>\n",
       "      <td>0.798670</td>\n",
       "      <td>0.616220</td>\n",
       "      <td>...</td>\n",
       "      <td>0.436448</td>\n",
       "      <td>0.807319</td>\n",
       "      <td>0.468079</td>\n",
       "      <td>0.474581</td>\n",
       "      <td>0.455604</td>\n",
       "      <td>0.473226</td>\n",
       "      <td>0.173221</td>\n",
       "      <td>0.703806</td>\n",
       "      <td>0.385992</td>\n",
       "      <td>0.438669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00006414</th>\n",
       "      <td>0.019880</td>\n",
       "      <td>0.036970</td>\n",
       "      <td>0.021106</td>\n",
       "      <td>0.026866</td>\n",
       "      <td>0.051002</td>\n",
       "      <td>0.038998</td>\n",
       "      <td>0.028451</td>\n",
       "      <td>0.029114</td>\n",
       "      <td>0.020005</td>\n",
       "      <td>0.043336</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026824</td>\n",
       "      <td>0.018624</td>\n",
       "      <td>0.044337</td>\n",
       "      <td>0.022170</td>\n",
       "      <td>0.028007</td>\n",
       "      <td>0.035340</td>\n",
       "      <td>0.038400</td>\n",
       "      <td>0.033976</td>\n",
       "      <td>0.084529</td>\n",
       "      <td>0.069691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27657283</th>\n",
       "      <td>0.108293</td>\n",
       "      <td>0.072173</td>\n",
       "      <td>0.068453</td>\n",
       "      <td>0.106587</td>\n",
       "      <td>0.072691</td>\n",
       "      <td>0.093206</td>\n",
       "      <td>0.119652</td>\n",
       "      <td>0.081737</td>\n",
       "      <td>0.082157</td>\n",
       "      <td>0.114968</td>\n",
       "      <td>...</td>\n",
       "      <td>0.144649</td>\n",
       "      <td>0.068423</td>\n",
       "      <td>0.083626</td>\n",
       "      <td>0.052279</td>\n",
       "      <td>0.054980</td>\n",
       "      <td>0.118309</td>\n",
       "      <td>0.064773</td>\n",
       "      <td>0.051555</td>\n",
       "      <td>0.219870</td>\n",
       "      <td>0.126036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27661264</th>\n",
       "      <td>0.359134</td>\n",
       "      <td>0.337503</td>\n",
       "      <td>0.233228</td>\n",
       "      <td>0.584919</td>\n",
       "      <td>0.343895</td>\n",
       "      <td>0.469068</td>\n",
       "      <td>0.506269</td>\n",
       "      <td>0.609027</td>\n",
       "      <td>0.360137</td>\n",
       "      <td>0.374381</td>\n",
       "      <td>...</td>\n",
       "      <td>0.364477</td>\n",
       "      <td>0.355658</td>\n",
       "      <td>0.384973</td>\n",
       "      <td>0.356396</td>\n",
       "      <td>0.352938</td>\n",
       "      <td>0.434815</td>\n",
       "      <td>0.371330</td>\n",
       "      <td>0.563619</td>\n",
       "      <td>0.529970</td>\n",
       "      <td>0.117175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27662379</th>\n",
       "      <td>0.008482</td>\n",
       "      <td>0.008813</td>\n",
       "      <td>0.008864</td>\n",
       "      <td>0.010997</td>\n",
       "      <td>0.024967</td>\n",
       "      <td>0.014811</td>\n",
       "      <td>0.020630</td>\n",
       "      <td>0.017429</td>\n",
       "      <td>0.014848</td>\n",
       "      <td>0.012520</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034058</td>\n",
       "      <td>0.009198</td>\n",
       "      <td>0.012508</td>\n",
       "      <td>0.014009</td>\n",
       "      <td>0.007969</td>\n",
       "      <td>0.010037</td>\n",
       "      <td>0.019814</td>\n",
       "      <td>0.014395</td>\n",
       "      <td>0.030977</td>\n",
       "      <td>0.025664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27662877</th>\n",
       "      <td>0.022540</td>\n",
       "      <td>0.018977</td>\n",
       "      <td>0.013427</td>\n",
       "      <td>0.031917</td>\n",
       "      <td>0.035480</td>\n",
       "      <td>0.047936</td>\n",
       "      <td>0.035445</td>\n",
       "      <td>0.027623</td>\n",
       "      <td>0.032830</td>\n",
       "      <td>0.087929</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034659</td>\n",
       "      <td>0.030906</td>\n",
       "      <td>0.027370</td>\n",
       "      <td>0.027446</td>\n",
       "      <td>0.017592</td>\n",
       "      <td>0.028750</td>\n",
       "      <td>0.036526</td>\n",
       "      <td>0.035873</td>\n",
       "      <td>0.048063</td>\n",
       "      <td>0.095890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg27665659</th>\n",
       "      <td>0.030471</td>\n",
       "      <td>0.023773</td>\n",
       "      <td>0.026649</td>\n",
       "      <td>0.029192</td>\n",
       "      <td>0.040387</td>\n",
       "      <td>0.028429</td>\n",
       "      <td>0.028926</td>\n",
       "      <td>0.020874</td>\n",
       "      <td>0.020866</td>\n",
       "      <td>0.034370</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043481</td>\n",
       "      <td>0.032193</td>\n",
       "      <td>0.030835</td>\n",
       "      <td>0.027944</td>\n",
       "      <td>0.029019</td>\n",
       "      <td>0.036316</td>\n",
       "      <td>0.033200</td>\n",
       "      <td>0.030438</td>\n",
       "      <td>0.040718</td>\n",
       "      <td>0.033083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27545 rows Ã— 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM1196428  GSM1196429  GSM1196430  GSM1196431  GSM1196432  \\\n",
       "ID                                                                       \n",
       "cg00000292    0.853864    0.873688    0.835080    0.598675    0.770182   \n",
       "cg00002426    0.183064    0.173539    0.205998    0.209957    0.176864   \n",
       "cg00003994    0.146456    0.123106    0.045628    0.081798    0.190403   \n",
       "cg00005847    0.855151    0.811072    0.830661    0.512991    0.738095   \n",
       "cg00006414    0.019880    0.036970    0.021106    0.026866    0.051002   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "cg27657283    0.108293    0.072173    0.068453    0.106587    0.072691   \n",
       "cg27661264    0.359134    0.337503    0.233228    0.584919    0.343895   \n",
       "cg27662379    0.008482    0.008813    0.008864    0.010997    0.024967   \n",
       "cg27662877    0.022540    0.018977    0.013427    0.031917    0.035480   \n",
       "cg27665659    0.030471    0.023773    0.026649    0.029192    0.040387   \n",
       "\n",
       "            GSM1196433  GSM1196434  GSM1196435  GSM1196436  GSM1196437  ...  \\\n",
       "ID                                                                      ...   \n",
       "cg00000292    0.435826    0.779509    0.832013    0.787016    0.727145  ...   \n",
       "cg00002426    0.227453    0.194177    0.214679    0.097024    0.366682  ...   \n",
       "cg00003994    0.100151    0.060844    0.056040    0.060491    0.219284  ...   \n",
       "cg00005847    0.367623    0.814609    0.681481    0.798670    0.616220  ...   \n",
       "cg00006414    0.038998    0.028451    0.029114    0.020005    0.043336  ...   \n",
       "...                ...         ...         ...         ...         ...  ...   \n",
       "cg27657283    0.093206    0.119652    0.081737    0.082157    0.114968  ...   \n",
       "cg27661264    0.469068    0.506269    0.609027    0.360137    0.374381  ...   \n",
       "cg27662379    0.014811    0.020630    0.017429    0.014848    0.012520  ...   \n",
       "cg27662877    0.047936    0.035445    0.027623    0.032830    0.087929  ...   \n",
       "cg27665659    0.028429    0.028926    0.020874    0.020866    0.034370  ...   \n",
       "\n",
       "            GSM1196501  GSM1196502  GSM1196503  GSM1196504  GSM1196505  \\\n",
       "ID                                                                       \n",
       "cg00000292    0.584722    0.407948    0.590603    0.567782    0.359869   \n",
       "cg00002426    0.217829    0.206476    0.169502    0.150175    0.228998   \n",
       "cg00003994    0.117100    0.054069    0.075588    0.076611    0.062579   \n",
       "cg00005847    0.436448    0.807319    0.468079    0.474581    0.455604   \n",
       "cg00006414    0.026824    0.018624    0.044337    0.022170    0.028007   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "cg27657283    0.144649    0.068423    0.083626    0.052279    0.054980   \n",
       "cg27661264    0.364477    0.355658    0.384973    0.356396    0.352938   \n",
       "cg27662379    0.034058    0.009198    0.012508    0.014009    0.007969   \n",
       "cg27662877    0.034659    0.030906    0.027370    0.027446    0.017592   \n",
       "cg27665659    0.043481    0.032193    0.030835    0.027944    0.029019   \n",
       "\n",
       "            GSM1196506  GSM1196507  GSM1196508  GSM1196509  GSM1196510  \n",
       "ID                                                                      \n",
       "cg00000292    0.653657    0.771832    0.349547    0.693256    0.438942  \n",
       "cg00002426    0.277141    0.248073    0.319916    0.258845    0.210580  \n",
       "cg00003994    0.188656    0.110269    0.058055    0.099765    0.113560  \n",
       "cg00005847    0.473226    0.173221    0.703806    0.385992    0.438669  \n",
       "cg00006414    0.035340    0.038400    0.033976    0.084529    0.069691  \n",
       "...                ...         ...         ...         ...         ...  \n",
       "cg27657283    0.118309    0.064773    0.051555    0.219870    0.126036  \n",
       "cg27661264    0.434815    0.371330    0.563619    0.529970    0.117175  \n",
       "cg27662379    0.010037    0.019814    0.014395    0.030977    0.025664  \n",
       "cg27662877    0.028750    0.036526    0.035873    0.048063    0.095890  \n",
       "cg27665659    0.036316    0.033200    0.030438    0.040718    0.033083  \n",
       "\n",
       "[27545 rows x 81 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'Name': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'IlmnStrand': ['TOP', 'TOP', 'TOP', 'BOT', 'BOT'], 'AddressA_ID': [990370.0, 6580397.0, 7150184.0, 4850717.0, 6980731.0], 'AlleleA_ProbeSeq': ['AAACATTAATTACCAACCACTCTTCCAAAAAACACTTACCATTAAAACCA', 'AATATAATAACATTACCTTACCCATCTTATAATCAAACCAAACAAAAACA', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCA', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACA', 'CTCAAAAACCAAACAAAACAAAACCCCAATACTAATCATTAATAAAATCA'], 'AddressB_ID': [6660678.0, 6100343.0, 7150392.0, 1260113.0, 4280093.0], 'AlleleB_ProbeSeq': ['AAACATTAATTACCAACCGCTCTTCCAAAAAACACTTACCATTAAAACCG', 'AATATAATAACATTACCTTACCCGTCTTATAATCAAACCAAACGAAAACG', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCG', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACG', 'CTCGAAAACCGAACAAAACAAAACCCCAATACTAATCGTTAATAAAATCG'], 'GenomeBuild': [36.0, 36.0, 36.0, 36.0, 36.0], 'Chr': ['16', '3', '7', '2', '7'], 'MapInfo': [28797601.0, 57718583.0, 15692387.0, 176737319.0, 148453770.0], 'Ploidy': ['diploid', 'diploid', 'diploid', 'diploid', 'diploid'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq'], 'SourceVersion': [36.1, 36.1, 36.1, 36.1, 36.1], 'SourceStrand': ['TOP', 'TOP', 'BOT', 'BOT', 'BOT'], 'SourceSeq': ['CGGCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCC', 'CGCTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATT', 'GGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTTCG', 'TACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAGCG', 'CTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGATCG'], 'TopGenomicSeq': ['TGGGGTGAGTGAGACCACGGGCCTCACCCCGGACCAAGTTAAGCGGAATCTGGAGAAATA[CG]GCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCCTCCTGCACCCCC', 'CCGCTGTCGACCAGCGCAGAATAATGCCACTTTTGATTGCAAAGTGCTATCAAGGAACCA[CG]CTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATTGTCCAGCGGCAT', 'GGTGGTGGTGGTGGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTT[CG]TTGGGGTATCCCGCGATTATGCAAGATGAGGAAGAAGTAGAGAGCTCGGGGTAAGACATA', 'CAGATAACTCAATACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAG[CG]CCATAAATTTGAAGGCCAATGATCGGTTTTCATGTAACGGGTGGTACTTCATACTGAAGT', 'GAACCGGCCCAGCTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGAT[CG]CGCCTGTGAATAGCCACTGCCCTCCAGCCTGGGCAACAGCCAGACCCCGTCTGTTTAATA'], 'Next_Base': ['T', 'T', 'T', 'C', 'C'], 'Color_Channel': ['Red', 'Red', 'Red', 'Grn', 'Grn'], 'TSS_Coordinate': [28797310.0, 57718214.0, 15692819.0, 176737051.0, 148454441.0], 'Gene_Strand': ['+', '+', '-', '+', '+'], 'Gene_ID': ['GeneID:487', 'GeneID:7871', 'GeneID:4223', 'GeneID:3232', 'GeneID:57541'], 'Symbol': ['ATP2A1', 'SLMAP', 'MEOX2', 'HOXD3', 'ZNF398'], 'Synonym': ['ATP2A; SERCA1;', 'SLAP; KIAA1601;', 'GAX; MOX2;', 'HOX4; HOX1D; HOX4A; Hox-4.1; MGC10470;', 'P51; P71; ZER6; KIAA1339;'], 'Accession': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'GID': ['GI:47132613', 'GI:56550042', 'GI:55956906', 'GI:23510372', 'GI:25777702'], 'Annotation': ['isoform a is encoded by transcript variant a; sarcoplasmic/endoplasmic reticulum calcium ATPase 1; calcium pump 1; SR Ca(2+)-ATPase 1; calcium-transporting ATPase sarcoplasmic reticulum type; fast twitch skeletal muscle isoform; endoplasmic reticulum class 1/2 Ca(2+) ATPase; go_component: membrane; go_component: integral to membrane; go_component: sarcoplasmic reticulum; go_component: smooth endoplasmic reticulum; go_function: ATP binding; go_function: hydrolase activity; go_function: nucleotide binding; go_function: calcium ion binding; go_function: magnesium ion binding; go_function: calcium-transporting ATPase activity; go_function: hydrolase activity; acting on acid anhydrides; catalyzing transmembrane movement of substances; go_process: metabolism; go_process: cation transport; go_process: proton transport; go_process: calcium ion transport; go_process: regulation of striated muscle contraction', 'Sarcolemmal-associated protein; go_component: integral to plasma membrane; go_component: smooth endoplasmic reticulum; go_function: unfolded protein binding; go_process: protein folding; go_process: muscle contraction', 'growth arrest-specific homeo box; go_component: nucleus; go_function: transcription factor activity; go_process: circulation; go_process: development; go_process: regulation of transcription; DNA-dependent', 'homeobox protein Hox-D3; Hox-4.1; mouse; homolog of; homeo box D3; go_component: nucleus; go_function: transcription factor activity; go_process: morphogenesis; go_process: regulation of transcription; DNA-dependent', 'isoform b is encoded by transcript variant 2; zinc finger DNA binding protein ZER6; zinc finger-estrogen receptor interaction; clone 6; zinc finger DNA binding protein p52/p71; go_component: nucleus; go_function: DNA binding; go_function: zinc ion binding; go_function: metal ion binding; go_function: transcriptional activator activity; go_process: transcription; go_process: regulation of transcription; DNA-dependent'], 'Product': ['ATPase; Ca++ transporting; fast twitch 1 isoform a', 'sarcolemma associated protein', 'mesenchyme homeo box 2', 'homeobox D3', 'zinc finger 398 isoform b'], 'Distance_to_TSS': [291.0, 369.0, 432.0, 268.0, 671.0], 'CPG_ISLAND': [True, True, True, False, True], 'CPG_ISLAND_LOCATIONS': ['16:28797486-28797825', '3:57716811-57718675', '7:15691512-15693551', nan, '7:148453584-148455804'], 'MIR_CPG_ISLAND': [nan, nan, nan, nan, nan], 'RANGE_GB': ['NC_000016.8', 'NC_000003.10', 'NC_000007.12', nan, 'NC_000007.12'], 'RANGE_START': [28797486.0, 57716811.0, 15691512.0, nan, 148453584.0], 'RANGE_END': [28797825.0, 57718675.0, 15693551.0, nan, 148455804.0], 'RANGE_STRAND': ['+', '+', '-', nan, '+'], 'GB_ACC': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'ORF': [487.0, 7871.0, 4223.0, 3232.0, 57541.0]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414', 'cg00007981', 'cg00008493', 'cg00008713', 'cg00009407', 'cg00010193', 'cg00011459', 'cg00012199', 'cg00012386', 'cg00012792', 'cg00013618', 'cg00014085', 'cg00014837', 'cg00015770', 'cg00016968', 'cg00019495']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'Name': ['cg00000292', 'cg00002426', 'cg00003994', 'cg00005847', 'cg00006414'], 'IlmnStrand': ['TOP', 'TOP', 'TOP', 'BOT', 'BOT'], 'AddressA_ID': [990370.0, 6580397.0, 7150184.0, 4850717.0, 6980731.0], 'AlleleA_ProbeSeq': ['AAACATTAATTACCAACCACTCTTCCAAAAAACACTTACCATTAAAACCA', 'AATATAATAACATTACCTTACCCATCTTATAATCAAACCAAACAAAAACA', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCA', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACA', 'CTCAAAAACCAAACAAAACAAAACCCCAATACTAATCATTAATAAAATCA'], 'AddressB_ID': [6660678.0, 6100343.0, 7150392.0, 1260113.0, 4280093.0], 'AlleleB_ProbeSeq': ['AAACATTAATTACCAACCGCTCTTCCAAAAAACACTTACCATTAAAACCG', 'AATATAATAACATTACCTTACCCGTCTTATAATCAAACCAAACGAAAACG', 'AATAATAATAATACCCCCTATAATACTAACTAACAAACATACCCTCTTCG', 'TACTATAATACACCCTATATTTAAAACACTAAACTTACCCCATTAAAACG', 'CTCGAAAACCGAACAAAACAAAACCCCAATACTAATCGTTAATAAAATCG'], 'GenomeBuild': [36.0, 36.0, 36.0, 36.0, 36.0], 'Chr': ['16', '3', '7', '2', '7'], 'MapInfo': [28797601.0, 57718583.0, 15692387.0, 176737319.0, 148453770.0], 'Ploidy': ['diploid', 'diploid', 'diploid', 'diploid', 'diploid'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq', 'NCBI:RefSeq'], 'SourceVersion': [36.1, 36.1, 36.1, 36.1, 36.1], 'SourceStrand': ['TOP', 'TOP', 'BOT', 'BOT', 'BOT'], 'SourceSeq': ['CGGCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCC', 'CGCTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATT', 'GGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTTCG', 'TACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAGCG', 'CTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGATCG'], 'TopGenomicSeq': ['TGGGGTGAGTGAGACCACGGGCCTCACCCCGGACCAAGTTAAGCGGAATCTGGAGAAATA[CG]GCCTCAATGGTAAGTGTCCCTTGGAAGAGCGGCTGGTAATTAATGCCCTCCTGCACCCCC', 'CCGCTGTCGACCAGCGCAGAATAATGCCACTTTTGATTGCAAAGTGCTATCAAGGAACCA[CG]CTCTCGTCTGGTTTGATCACAAGACGGGCAAGGTAATGTCACCACATTGTCCAGCGGCAT', 'GGTGGTGGTGGTGGTGGTGGTGGTGCCCCCTGTGATGCTGGCTGGCAAACATGCCCTCTT[CG]TTGGGGTATCCCGCGATTATGCAAGATGAGGAAGAAGTAGAGAGCTCGGGGTAAGACATA', 'CAGATAACTCAATACTGTAATGCACCCTGTATTTAAGGCACTGGGCTTGCCCCATTAAAG[CG]CCATAAATTTGAAGGCCAATGATCGGTTTTCATGTAACGGGTGGTACTTCATACTGAAGT', 'GAACCGGCCCAGCTCGGAAACCGAGCAGGGCAAAACCCCAGTGCTGATCGTTAGTGGGAT[CG]CGCCTGTGAATAGCCACTGCCCTCCAGCCTGGGCAACAGCCAGACCCCGTCTGTTTAATA'], 'Next_Base': ['T', 'T', 'T', 'C', 'C'], 'Color_Channel': ['Red', 'Red', 'Red', 'Grn', 'Grn'], 'TSS_Coordinate': [28797310.0, 57718214.0, 15692819.0, 176737051.0, 148454441.0], 'Gene_Strand': ['+', '+', '-', '+', '+'], 'Gene_ID': ['GeneID:487', 'GeneID:7871', 'GeneID:4223', 'GeneID:3232', 'GeneID:57541'], 'Symbol': ['ATP2A1', 'SLMAP', 'MEOX2', 'HOXD3', 'ZNF398'], 'Synonym': ['ATP2A; SERCA1;', 'SLAP; KIAA1601;', 'GAX; MOX2;', 'HOX4; HOX1D; HOX4A; Hox-4.1; MGC10470;', 'P51; P71; ZER6; KIAA1339;'], 'Accession': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'GID': ['GI:47132613', 'GI:56550042', 'GI:55956906', 'GI:23510372', 'GI:25777702'], 'Annotation': ['isoform a is encoded by transcript variant a; sarcoplasmic/endoplasmic reticulum calcium ATPase 1; calcium pump 1; SR Ca(2+)-ATPase 1; calcium-transporting ATPase sarcoplasmic reticulum type; fast twitch skeletal muscle isoform; endoplasmic reticulum class 1/2 Ca(2+) ATPase; go_component: membrane; go_component: integral to membrane; go_component: sarcoplasmic reticulum; go_component: smooth endoplasmic reticulum; go_function: ATP binding; go_function: hydrolase activity; go_function: nucleotide binding; go_function: calcium ion binding; go_function: magnesium ion binding; go_function: calcium-transporting ATPase activity; go_function: hydrolase activity; acting on acid anhydrides; catalyzing transmembrane movement of substances; go_process: metabolism; go_process: cation transport; go_process: proton transport; go_process: calcium ion transport; go_process: regulation of striated muscle contraction', 'Sarcolemmal-associated protein; go_component: integral to plasma membrane; go_component: smooth endoplasmic reticulum; go_function: unfolded protein binding; go_process: protein folding; go_process: muscle contraction', 'growth arrest-specific homeo box; go_component: nucleus; go_function: transcription factor activity; go_process: circulation; go_process: development; go_process: regulation of transcription; DNA-dependent', 'homeobox protein Hox-D3; Hox-4.1; mouse; homolog of; homeo box D3; go_component: nucleus; go_function: transcription factor activity; go_process: morphogenesis; go_process: regulation of transcription; DNA-dependent', 'isoform b is encoded by transcript variant 2; zinc finger DNA binding protein ZER6; zinc finger-estrogen receptor interaction; clone 6; zinc finger DNA binding protein p52/p71; go_component: nucleus; go_function: DNA binding; go_function: zinc ion binding; go_function: metal ion binding; go_function: transcriptional activator activity; go_process: transcription; go_process: regulation of transcription; DNA-dependent'], 'Product': ['ATPase; Ca++ transporting; fast twitch 1 isoform a', 'sarcolemma associated protein', 'mesenchyme homeo box 2', 'homeobox D3', 'zinc finger 398 isoform b'], 'Distance_to_TSS': [291.0, 369.0, 432.0, 268.0, 671.0], 'CPG_ISLAND': [True, True, True, False, True], 'CPG_ISLAND_LOCATIONS': ['16:28797486-28797825', '3:57716811-57718675', '7:15691512-15693551', nan, '7:148453584-148455804'], 'MIR_CPG_ISLAND': [nan, nan, nan, nan, nan], 'RANGE_GB': ['NC_000016.8', 'NC_000003.10', 'NC_000007.12', nan, 'NC_000007.12'], 'RANGE_START': [28797486.0, 57716811.0, 15691512.0, nan, 148453584.0], 'RANGE_END': [28797825.0, 57718675.0, 15693551.0, nan, 148455804.0], 'RANGE_STRAND': ['+', '+', '-', nan, '+'], 'GB_ACC': ['NM_173201.2', 'NM_007159.2', 'NM_005924.3', 'NM_006898.4', 'NM_020781.2'], 'ORF': [487.0, 7871.0, 4223.0, 3232.0, 57541.0]}\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Name', 'IlmnStrand', 'AddressA_ID', 'AlleleA_ProbeSeq',\n",
       "       'AddressB_ID', 'AlleleB_ProbeSeq', 'GenomeBuild', 'Chr', 'MapInfo',\n",
       "       'Ploidy', 'Species', 'Source', 'SourceVersion', 'SourceStrand',\n",
       "       'SourceSeq', 'TopGenomicSeq', 'Next_Base', 'Color_Channel',\n",
       "       'TSS_Coordinate', 'Gene_Strand', 'Gene_ID', 'Symbol', 'Synonym',\n",
       "       'Accession', 'GID', 'Annotation', 'Product', 'Distance_to_TSS',\n",
       "       'CPG_ISLAND', 'CPG_ISLAND_LOCATIONS', 'MIR_CPG_ISLAND', 'RANGE_GB',\n",
       "       'RANGE_START', 'RANGE_END', 'RANGE_STRAND', 'GB_ACC', 'ORF'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196428</th>\n",
       "      <th>GSM1196429</th>\n",
       "      <th>GSM1196430</th>\n",
       "      <th>GSM1196431</th>\n",
       "      <th>GSM1196432</th>\n",
       "      <th>GSM1196433</th>\n",
       "      <th>GSM1196434</th>\n",
       "      <th>GSM1196435</th>\n",
       "      <th>GSM1196436</th>\n",
       "      <th>GSM1196437</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196501</th>\n",
       "      <th>GSM1196502</th>\n",
       "      <th>GSM1196503</th>\n",
       "      <th>GSM1196504</th>\n",
       "      <th>GSM1196505</th>\n",
       "      <th>GSM1196506</th>\n",
       "      <th>GSM1196507</th>\n",
       "      <th>GSM1196508</th>\n",
       "      <th>GSM1196509</th>\n",
       "      <th>GSM1196510</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39873</th>\n",
       "      <td>0.673936</td>\n",
       "      <td>0.767293</td>\n",
       "      <td>0.462837</td>\n",
       "      <td>0.723856</td>\n",
       "      <td>0.762069</td>\n",
       "      <td>0.667737</td>\n",
       "      <td>0.626288</td>\n",
       "      <td>0.565551</td>\n",
       "      <td>0.443497</td>\n",
       "      <td>0.342631</td>\n",
       "      <td>...</td>\n",
       "      <td>0.513662</td>\n",
       "      <td>0.660481</td>\n",
       "      <td>0.552440</td>\n",
       "      <td>0.520529</td>\n",
       "      <td>0.479534</td>\n",
       "      <td>0.420159</td>\n",
       "      <td>0.610865</td>\n",
       "      <td>0.699102</td>\n",
       "      <td>0.693876</td>\n",
       "      <td>0.663477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39874</th>\n",
       "      <td>0.021972</td>\n",
       "      <td>0.027994</td>\n",
       "      <td>0.021949</td>\n",
       "      <td>0.035239</td>\n",
       "      <td>0.043132</td>\n",
       "      <td>0.030231</td>\n",
       "      <td>0.040192</td>\n",
       "      <td>0.025203</td>\n",
       "      <td>0.034531</td>\n",
       "      <td>0.049580</td>\n",
       "      <td>...</td>\n",
       "      <td>0.069800</td>\n",
       "      <td>0.023893</td>\n",
       "      <td>0.033235</td>\n",
       "      <td>0.045726</td>\n",
       "      <td>0.035371</td>\n",
       "      <td>0.036387</td>\n",
       "      <td>0.038013</td>\n",
       "      <td>0.061310</td>\n",
       "      <td>0.068406</td>\n",
       "      <td>0.080097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39875</th>\n",
       "      <td>0.257827</td>\n",
       "      <td>0.335861</td>\n",
       "      <td>0.430710</td>\n",
       "      <td>0.261546</td>\n",
       "      <td>0.417659</td>\n",
       "      <td>0.235851</td>\n",
       "      <td>0.444144</td>\n",
       "      <td>0.425444</td>\n",
       "      <td>0.279365</td>\n",
       "      <td>0.418262</td>\n",
       "      <td>...</td>\n",
       "      <td>0.397831</td>\n",
       "      <td>0.425723</td>\n",
       "      <td>0.161505</td>\n",
       "      <td>0.417688</td>\n",
       "      <td>0.279023</td>\n",
       "      <td>0.414954</td>\n",
       "      <td>0.339570</td>\n",
       "      <td>0.422286</td>\n",
       "      <td>0.398828</td>\n",
       "      <td>0.105795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39877</th>\n",
       "      <td>0.029100</td>\n",
       "      <td>0.029369</td>\n",
       "      <td>0.020777</td>\n",
       "      <td>0.038653</td>\n",
       "      <td>0.048058</td>\n",
       "      <td>0.036818</td>\n",
       "      <td>0.031426</td>\n",
       "      <td>0.040999</td>\n",
       "      <td>0.031261</td>\n",
       "      <td>0.044107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.061673</td>\n",
       "      <td>0.030315</td>\n",
       "      <td>0.035297</td>\n",
       "      <td>0.030667</td>\n",
       "      <td>0.038152</td>\n",
       "      <td>0.039511</td>\n",
       "      <td>0.034959</td>\n",
       "      <td>0.046571</td>\n",
       "      <td>0.067835</td>\n",
       "      <td>0.083983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39878</th>\n",
       "      <td>0.037687</td>\n",
       "      <td>0.036334</td>\n",
       "      <td>0.041640</td>\n",
       "      <td>0.041173</td>\n",
       "      <td>0.052805</td>\n",
       "      <td>0.061906</td>\n",
       "      <td>0.048494</td>\n",
       "      <td>0.041754</td>\n",
       "      <td>0.048721</td>\n",
       "      <td>0.049560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.059706</td>\n",
       "      <td>0.043102</td>\n",
       "      <td>0.038224</td>\n",
       "      <td>0.047865</td>\n",
       "      <td>0.049129</td>\n",
       "      <td>0.050167</td>\n",
       "      <td>0.038020</td>\n",
       "      <td>0.059438</td>\n",
       "      <td>0.068282</td>\n",
       "      <td>0.077149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-D3</th>\n",
       "      <td>0.086622</td>\n",
       "      <td>0.293496</td>\n",
       "      <td>0.511238</td>\n",
       "      <td>0.346532</td>\n",
       "      <td>0.331784</td>\n",
       "      <td>0.211769</td>\n",
       "      <td>0.771208</td>\n",
       "      <td>0.387434</td>\n",
       "      <td>0.370193</td>\n",
       "      <td>0.178503</td>\n",
       "      <td>...</td>\n",
       "      <td>0.195070</td>\n",
       "      <td>0.139705</td>\n",
       "      <td>0.209928</td>\n",
       "      <td>0.274113</td>\n",
       "      <td>0.115736</td>\n",
       "      <td>0.183836</td>\n",
       "      <td>0.199735</td>\n",
       "      <td>0.245198</td>\n",
       "      <td>0.204434</td>\n",
       "      <td>0.156846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-H2</th>\n",
       "      <td>0.018871</td>\n",
       "      <td>0.033459</td>\n",
       "      <td>0.017983</td>\n",
       "      <td>0.032419</td>\n",
       "      <td>0.056566</td>\n",
       "      <td>0.035238</td>\n",
       "      <td>0.029841</td>\n",
       "      <td>0.019733</td>\n",
       "      <td>0.031213</td>\n",
       "      <td>0.032566</td>\n",
       "      <td>...</td>\n",
       "      <td>0.039907</td>\n",
       "      <td>0.023451</td>\n",
       "      <td>0.037152</td>\n",
       "      <td>0.022379</td>\n",
       "      <td>0.021213</td>\n",
       "      <td>0.031702</td>\n",
       "      <td>0.034143</td>\n",
       "      <td>0.051531</td>\n",
       "      <td>0.052094</td>\n",
       "      <td>0.045951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hfl-B5</th>\n",
       "      <td>0.025806</td>\n",
       "      <td>0.026141</td>\n",
       "      <td>0.023396</td>\n",
       "      <td>0.033722</td>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.024440</td>\n",
       "      <td>0.029792</td>\n",
       "      <td>0.029799</td>\n",
       "      <td>0.024806</td>\n",
       "      <td>0.028791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037977</td>\n",
       "      <td>0.023297</td>\n",
       "      <td>0.023845</td>\n",
       "      <td>0.022994</td>\n",
       "      <td>0.027307</td>\n",
       "      <td>0.030210</td>\n",
       "      <td>0.029699</td>\n",
       "      <td>0.036275</td>\n",
       "      <td>0.039542</td>\n",
       "      <td>0.050389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mimitin</th>\n",
       "      <td>0.795290</td>\n",
       "      <td>0.764163</td>\n",
       "      <td>0.795372</td>\n",
       "      <td>0.662529</td>\n",
       "      <td>0.704436</td>\n",
       "      <td>0.654050</td>\n",
       "      <td>0.785667</td>\n",
       "      <td>0.690498</td>\n",
       "      <td>0.790545</td>\n",
       "      <td>0.636648</td>\n",
       "      <td>...</td>\n",
       "      <td>0.640861</td>\n",
       "      <td>0.639229</td>\n",
       "      <td>0.648629</td>\n",
       "      <td>0.687194</td>\n",
       "      <td>0.611046</td>\n",
       "      <td>0.588422</td>\n",
       "      <td>0.551924</td>\n",
       "      <td>0.418062</td>\n",
       "      <td>0.579748</td>\n",
       "      <td>0.585062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>raptor</th>\n",
       "      <td>0.089944</td>\n",
       "      <td>0.029531</td>\n",
       "      <td>0.032989</td>\n",
       "      <td>0.047693</td>\n",
       "      <td>0.045761</td>\n",
       "      <td>0.064970</td>\n",
       "      <td>0.068787</td>\n",
       "      <td>0.041650</td>\n",
       "      <td>0.042405</td>\n",
       "      <td>0.055766</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078578</td>\n",
       "      <td>0.038967</td>\n",
       "      <td>0.043212</td>\n",
       "      <td>0.030947</td>\n",
       "      <td>0.040653</td>\n",
       "      <td>0.056151</td>\n",
       "      <td>0.037542</td>\n",
       "      <td>0.046660</td>\n",
       "      <td>0.125524</td>\n",
       "      <td>0.101505</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14473 rows Ã— 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1196428  GSM1196429  GSM1196430  GSM1196431  GSM1196432  \\\n",
       "Gene                                                                  \n",
       "39873      0.673936    0.767293    0.462837    0.723856    0.762069   \n",
       "39874      0.021972    0.027994    0.021949    0.035239    0.043132   \n",
       "39875      0.257827    0.335861    0.430710    0.261546    0.417659   \n",
       "39877      0.029100    0.029369    0.020777    0.038653    0.048058   \n",
       "39878      0.037687    0.036334    0.041640    0.041173    0.052805   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "hCAP-D3    0.086622    0.293496    0.511238    0.346532    0.331784   \n",
       "hCAP-H2    0.018871    0.033459    0.017983    0.032419    0.056566   \n",
       "hfl-B5     0.025806    0.026141    0.023396    0.033722    0.041915   \n",
       "mimitin    0.795290    0.764163    0.795372    0.662529    0.704436   \n",
       "raptor     0.089944    0.029531    0.032989    0.047693    0.045761   \n",
       "\n",
       "         GSM1196433  GSM1196434  GSM1196435  GSM1196436  GSM1196437  ...  \\\n",
       "Gene                                                                 ...   \n",
       "39873      0.667737    0.626288    0.565551    0.443497    0.342631  ...   \n",
       "39874      0.030231    0.040192    0.025203    0.034531    0.049580  ...   \n",
       "39875      0.235851    0.444144    0.425444    0.279365    0.418262  ...   \n",
       "39877      0.036818    0.031426    0.040999    0.031261    0.044107  ...   \n",
       "39878      0.061906    0.048494    0.041754    0.048721    0.049560  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "hCAP-D3    0.211769    0.771208    0.387434    0.370193    0.178503  ...   \n",
       "hCAP-H2    0.035238    0.029841    0.019733    0.031213    0.032566  ...   \n",
       "hfl-B5     0.024440    0.029792    0.029799    0.024806    0.028791  ...   \n",
       "mimitin    0.654050    0.785667    0.690498    0.790545    0.636648  ...   \n",
       "raptor     0.064970    0.068787    0.041650    0.042405    0.055766  ...   \n",
       "\n",
       "         GSM1196501  GSM1196502  GSM1196503  GSM1196504  GSM1196505  \\\n",
       "Gene                                                                  \n",
       "39873      0.513662    0.660481    0.552440    0.520529    0.479534   \n",
       "39874      0.069800    0.023893    0.033235    0.045726    0.035371   \n",
       "39875      0.397831    0.425723    0.161505    0.417688    0.279023   \n",
       "39877      0.061673    0.030315    0.035297    0.030667    0.038152   \n",
       "39878      0.059706    0.043102    0.038224    0.047865    0.049129   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "hCAP-D3    0.195070    0.139705    0.209928    0.274113    0.115736   \n",
       "hCAP-H2    0.039907    0.023451    0.037152    0.022379    0.021213   \n",
       "hfl-B5     0.037977    0.023297    0.023845    0.022994    0.027307   \n",
       "mimitin    0.640861    0.639229    0.648629    0.687194    0.611046   \n",
       "raptor     0.078578    0.038967    0.043212    0.030947    0.040653   \n",
       "\n",
       "         GSM1196506  GSM1196507  GSM1196508  GSM1196509  GSM1196510  \n",
       "Gene                                                                 \n",
       "39873      0.420159    0.610865    0.699102    0.693876    0.663477  \n",
       "39874      0.036387    0.038013    0.061310    0.068406    0.080097  \n",
       "39875      0.414954    0.339570    0.422286    0.398828    0.105795  \n",
       "39877      0.039511    0.034959    0.046571    0.067835    0.083983  \n",
       "39878      0.050167    0.038020    0.059438    0.068282    0.077149  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "hCAP-D3    0.183836    0.199735    0.245198    0.204434    0.156846  \n",
       "hCAP-H2    0.031702    0.034143    0.051531    0.052094    0.045951  \n",
       "hfl-B5     0.030210    0.029699    0.036275    0.039542    0.050389  \n",
       "mimitin    0.588422    0.551924    0.418062    0.579748    0.585062  \n",
       "raptor     0.056151    0.037542    0.046660    0.125524    0.101505  \n",
       "\n",
       "[14473 rows x 81 columns]"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196428</th>\n",
       "      <th>GSM1196429</th>\n",
       "      <th>GSM1196430</th>\n",
       "      <th>GSM1196431</th>\n",
       "      <th>GSM1196432</th>\n",
       "      <th>GSM1196433</th>\n",
       "      <th>GSM1196434</th>\n",
       "      <th>GSM1196435</th>\n",
       "      <th>GSM1196436</th>\n",
       "      <th>GSM1196437</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196501</th>\n",
       "      <th>GSM1196502</th>\n",
       "      <th>GSM1196503</th>\n",
       "      <th>GSM1196504</th>\n",
       "      <th>GSM1196505</th>\n",
       "      <th>GSM1196506</th>\n",
       "      <th>GSM1196507</th>\n",
       "      <th>GSM1196508</th>\n",
       "      <th>GSM1196509</th>\n",
       "      <th>GSM1196510</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.940233</td>\n",
       "      <td>0.955381</td>\n",
       "      <td>0.938320</td>\n",
       "      <td>0.872892</td>\n",
       "      <td>0.916779</td>\n",
       "      <td>0.856238</td>\n",
       "      <td>0.955121</td>\n",
       "      <td>0.934917</td>\n",
       "      <td>0.936212</td>\n",
       "      <td>0.921644</td>\n",
       "      <td>...</td>\n",
       "      <td>0.935269</td>\n",
       "      <td>0.913290</td>\n",
       "      <td>0.934775</td>\n",
       "      <td>0.876953</td>\n",
       "      <td>0.889109</td>\n",
       "      <td>0.939685</td>\n",
       "      <td>0.815038</td>\n",
       "      <td>0.561779</td>\n",
       "      <td>0.883203</td>\n",
       "      <td>0.890838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.749546</td>\n",
       "      <td>0.245808</td>\n",
       "      <td>0.694354</td>\n",
       "      <td>0.460610</td>\n",
       "      <td>0.506125</td>\n",
       "      <td>0.281712</td>\n",
       "      <td>0.565541</td>\n",
       "      <td>0.474934</td>\n",
       "      <td>0.689960</td>\n",
       "      <td>0.745397</td>\n",
       "      <td>...</td>\n",
       "      <td>0.437034</td>\n",
       "      <td>0.584193</td>\n",
       "      <td>0.562674</td>\n",
       "      <td>0.578834</td>\n",
       "      <td>0.324707</td>\n",
       "      <td>0.527137</td>\n",
       "      <td>0.561500</td>\n",
       "      <td>0.482367</td>\n",
       "      <td>0.434855</td>\n",
       "      <td>0.641983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.790944</td>\n",
       "      <td>0.862709</td>\n",
       "      <td>0.828572</td>\n",
       "      <td>0.857309</td>\n",
       "      <td>0.823975</td>\n",
       "      <td>0.823414</td>\n",
       "      <td>0.851734</td>\n",
       "      <td>0.866243</td>\n",
       "      <td>0.856738</td>\n",
       "      <td>0.832075</td>\n",
       "      <td>...</td>\n",
       "      <td>0.796702</td>\n",
       "      <td>0.762879</td>\n",
       "      <td>0.781257</td>\n",
       "      <td>0.700352</td>\n",
       "      <td>0.744846</td>\n",
       "      <td>0.712289</td>\n",
       "      <td>0.735023</td>\n",
       "      <td>0.824379</td>\n",
       "      <td>0.797928</td>\n",
       "      <td>0.724440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.374478</td>\n",
       "      <td>0.458769</td>\n",
       "      <td>0.362774</td>\n",
       "      <td>0.277094</td>\n",
       "      <td>0.565643</td>\n",
       "      <td>0.452039</td>\n",
       "      <td>0.435517</td>\n",
       "      <td>0.336927</td>\n",
       "      <td>0.447911</td>\n",
       "      <td>0.367933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.356745</td>\n",
       "      <td>0.279629</td>\n",
       "      <td>0.240643</td>\n",
       "      <td>0.425051</td>\n",
       "      <td>0.302600</td>\n",
       "      <td>0.342275</td>\n",
       "      <td>0.300820</td>\n",
       "      <td>0.312059</td>\n",
       "      <td>0.343013</td>\n",
       "      <td>0.261995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>0.634482</td>\n",
       "      <td>0.861004</td>\n",
       "      <td>0.768237</td>\n",
       "      <td>0.732757</td>\n",
       "      <td>0.848816</td>\n",
       "      <td>0.802036</td>\n",
       "      <td>0.815582</td>\n",
       "      <td>0.811549</td>\n",
       "      <td>0.768936</td>\n",
       "      <td>0.744320</td>\n",
       "      <td>...</td>\n",
       "      <td>0.766266</td>\n",
       "      <td>0.729659</td>\n",
       "      <td>0.790821</td>\n",
       "      <td>0.782283</td>\n",
       "      <td>0.760710</td>\n",
       "      <td>0.810943</td>\n",
       "      <td>0.830674</td>\n",
       "      <td>0.865127</td>\n",
       "      <td>0.758200</td>\n",
       "      <td>0.620545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWINT</th>\n",
       "      <td>0.031971</td>\n",
       "      <td>0.037855</td>\n",
       "      <td>0.033131</td>\n",
       "      <td>0.036843</td>\n",
       "      <td>0.043600</td>\n",
       "      <td>0.045227</td>\n",
       "      <td>0.041920</td>\n",
       "      <td>0.032721</td>\n",
       "      <td>0.032561</td>\n",
       "      <td>0.053831</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053125</td>\n",
       "      <td>0.034923</td>\n",
       "      <td>0.031704</td>\n",
       "      <td>0.028981</td>\n",
       "      <td>0.032775</td>\n",
       "      <td>0.031853</td>\n",
       "      <td>0.032822</td>\n",
       "      <td>0.043026</td>\n",
       "      <td>0.066233</td>\n",
       "      <td>0.072853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDA</th>\n",
       "      <td>0.468787</td>\n",
       "      <td>0.468326</td>\n",
       "      <td>0.457744</td>\n",
       "      <td>0.577638</td>\n",
       "      <td>0.449071</td>\n",
       "      <td>0.473941</td>\n",
       "      <td>0.508964</td>\n",
       "      <td>0.514308</td>\n",
       "      <td>0.402810</td>\n",
       "      <td>0.496837</td>\n",
       "      <td>...</td>\n",
       "      <td>0.428336</td>\n",
       "      <td>0.532234</td>\n",
       "      <td>0.419962</td>\n",
       "      <td>0.428588</td>\n",
       "      <td>0.465011</td>\n",
       "      <td>0.403560</td>\n",
       "      <td>0.539247</td>\n",
       "      <td>0.501775</td>\n",
       "      <td>0.458448</td>\n",
       "      <td>0.518576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.029696</td>\n",
       "      <td>0.019655</td>\n",
       "      <td>0.029563</td>\n",
       "      <td>0.034270</td>\n",
       "      <td>0.032857</td>\n",
       "      <td>0.035739</td>\n",
       "      <td>0.027779</td>\n",
       "      <td>0.023625</td>\n",
       "      <td>0.027304</td>\n",
       "      <td>0.023680</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040760</td>\n",
       "      <td>0.027923</td>\n",
       "      <td>0.029210</td>\n",
       "      <td>0.032056</td>\n",
       "      <td>0.032080</td>\n",
       "      <td>0.024537</td>\n",
       "      <td>0.036141</td>\n",
       "      <td>0.028701</td>\n",
       "      <td>0.040045</td>\n",
       "      <td>0.033771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.079243</td>\n",
       "      <td>0.072206</td>\n",
       "      <td>0.066787</td>\n",
       "      <td>0.174444</td>\n",
       "      <td>0.439872</td>\n",
       "      <td>0.073742</td>\n",
       "      <td>0.225078</td>\n",
       "      <td>0.072467</td>\n",
       "      <td>0.072189</td>\n",
       "      <td>0.117451</td>\n",
       "      <td>...</td>\n",
       "      <td>0.097018</td>\n",
       "      <td>0.065814</td>\n",
       "      <td>0.082594</td>\n",
       "      <td>0.078288</td>\n",
       "      <td>0.085644</td>\n",
       "      <td>0.085534</td>\n",
       "      <td>0.130871</td>\n",
       "      <td>0.056252</td>\n",
       "      <td>0.089391</td>\n",
       "      <td>0.095180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.022379</td>\n",
       "      <td>0.044275</td>\n",
       "      <td>0.047593</td>\n",
       "      <td>0.045351</td>\n",
       "      <td>0.053067</td>\n",
       "      <td>0.048845</td>\n",
       "      <td>0.054456</td>\n",
       "      <td>0.045733</td>\n",
       "      <td>0.039186</td>\n",
       "      <td>0.040406</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060621</td>\n",
       "      <td>0.028046</td>\n",
       "      <td>0.042696</td>\n",
       "      <td>0.040194</td>\n",
       "      <td>0.036972</td>\n",
       "      <td>0.053363</td>\n",
       "      <td>0.039618</td>\n",
       "      <td>0.035059</td>\n",
       "      <td>0.090369</td>\n",
       "      <td>0.079189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11168 rows Ã— 81 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM1196428  GSM1196429  GSM1196430  GSM1196431  GSM1196432  \\\n",
       "A1BG      0.940233    0.955381    0.938320    0.872892    0.916779   \n",
       "A2M       0.749546    0.245808    0.694354    0.460610    0.506125   \n",
       "A2ML1     0.790944    0.862709    0.828572    0.857309    0.823975   \n",
       "A4GALT    0.374478    0.458769    0.362774    0.277094    0.565643   \n",
       "A4GNT     0.634482    0.861004    0.768237    0.732757    0.848816   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZWINT     0.031971    0.037855    0.033131    0.036843    0.043600   \n",
       "ZXDA      0.468787    0.468326    0.457744    0.577638    0.449071   \n",
       "ZYX       0.029696    0.019655    0.029563    0.034270    0.032857   \n",
       "ZZEF1     0.079243    0.072206    0.066787    0.174444    0.439872   \n",
       "ZZZ3      0.022379    0.044275    0.047593    0.045351    0.053067   \n",
       "\n",
       "        GSM1196433  GSM1196434  GSM1196435  GSM1196436  GSM1196437  ...  \\\n",
       "A1BG      0.856238    0.955121    0.934917    0.936212    0.921644  ...   \n",
       "A2M       0.281712    0.565541    0.474934    0.689960    0.745397  ...   \n",
       "A2ML1     0.823414    0.851734    0.866243    0.856738    0.832075  ...   \n",
       "A4GALT    0.452039    0.435517    0.336927    0.447911    0.367933  ...   \n",
       "A4GNT     0.802036    0.815582    0.811549    0.768936    0.744320  ...   \n",
       "...            ...         ...         ...         ...         ...  ...   \n",
       "ZWINT     0.045227    0.041920    0.032721    0.032561    0.053831  ...   \n",
       "ZXDA      0.473941    0.508964    0.514308    0.402810    0.496837  ...   \n",
       "ZYX       0.035739    0.027779    0.023625    0.027304    0.023680  ...   \n",
       "ZZEF1     0.073742    0.225078    0.072467    0.072189    0.117451  ...   \n",
       "ZZZ3      0.048845    0.054456    0.045733    0.039186    0.040406  ...   \n",
       "\n",
       "        GSM1196501  GSM1196502  GSM1196503  GSM1196504  GSM1196505  \\\n",
       "A1BG      0.935269    0.913290    0.934775    0.876953    0.889109   \n",
       "A2M       0.437034    0.584193    0.562674    0.578834    0.324707   \n",
       "A2ML1     0.796702    0.762879    0.781257    0.700352    0.744846   \n",
       "A4GALT    0.356745    0.279629    0.240643    0.425051    0.302600   \n",
       "A4GNT     0.766266    0.729659    0.790821    0.782283    0.760710   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZWINT     0.053125    0.034923    0.031704    0.028981    0.032775   \n",
       "ZXDA      0.428336    0.532234    0.419962    0.428588    0.465011   \n",
       "ZYX       0.040760    0.027923    0.029210    0.032056    0.032080   \n",
       "ZZEF1     0.097018    0.065814    0.082594    0.078288    0.085644   \n",
       "ZZZ3      0.060621    0.028046    0.042696    0.040194    0.036972   \n",
       "\n",
       "        GSM1196506  GSM1196507  GSM1196508  GSM1196509  GSM1196510  \n",
       "A1BG      0.939685    0.815038    0.561779    0.883203    0.890838  \n",
       "A2M       0.527137    0.561500    0.482367    0.434855    0.641983  \n",
       "A2ML1     0.712289    0.735023    0.824379    0.797928    0.724440  \n",
       "A4GALT    0.342275    0.300820    0.312059    0.343013    0.261995  \n",
       "A4GNT     0.810943    0.830674    0.865127    0.758200    0.620545  \n",
       "...            ...         ...         ...         ...         ...  \n",
       "ZWINT     0.031853    0.032822    0.043026    0.066233    0.072853  \n",
       "ZXDA      0.403560    0.539247    0.501775    0.458448    0.518576  \n",
       "ZYX       0.024537    0.036141    0.028701    0.040045    0.033771  \n",
       "ZZEF1     0.085534    0.130871    0.056252    0.089391    0.095180  \n",
       "ZZZ3      0.053363    0.039618    0.035059    0.090369    0.079189  \n",
       "\n",
       "[11168 rows x 81 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AADACL2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN2</th>\n",
       "      <th>ZSCAN4</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1196428</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.940233</td>\n",
       "      <td>0.749546</td>\n",
       "      <td>0.790944</td>\n",
       "      <td>0.374478</td>\n",
       "      <td>0.634482</td>\n",
       "      <td>0.052668</td>\n",
       "      <td>0.217289</td>\n",
       "      <td>0.830051</td>\n",
       "      <td>0.672389</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040573</td>\n",
       "      <td>0.810611</td>\n",
       "      <td>0.046088</td>\n",
       "      <td>0.035631</td>\n",
       "      <td>0.037018</td>\n",
       "      <td>0.031971</td>\n",
       "      <td>0.468787</td>\n",
       "      <td>0.029696</td>\n",
       "      <td>0.079243</td>\n",
       "      <td>0.022379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196429</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.955381</td>\n",
       "      <td>0.245808</td>\n",
       "      <td>0.862709</td>\n",
       "      <td>0.458769</td>\n",
       "      <td>0.861004</td>\n",
       "      <td>0.062860</td>\n",
       "      <td>0.070667</td>\n",
       "      <td>0.920872</td>\n",
       "      <td>0.661086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054629</td>\n",
       "      <td>0.947878</td>\n",
       "      <td>0.069931</td>\n",
       "      <td>0.035145</td>\n",
       "      <td>0.036839</td>\n",
       "      <td>0.037855</td>\n",
       "      <td>0.468326</td>\n",
       "      <td>0.019655</td>\n",
       "      <td>0.072206</td>\n",
       "      <td>0.044275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196430</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.938320</td>\n",
       "      <td>0.694354</td>\n",
       "      <td>0.828572</td>\n",
       "      <td>0.362774</td>\n",
       "      <td>0.768237</td>\n",
       "      <td>0.062411</td>\n",
       "      <td>0.085289</td>\n",
       "      <td>0.801509</td>\n",
       "      <td>0.542964</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040528</td>\n",
       "      <td>0.850798</td>\n",
       "      <td>0.063188</td>\n",
       "      <td>0.038473</td>\n",
       "      <td>0.043276</td>\n",
       "      <td>0.033131</td>\n",
       "      <td>0.457744</td>\n",
       "      <td>0.029563</td>\n",
       "      <td>0.066787</td>\n",
       "      <td>0.047593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196431</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.872892</td>\n",
       "      <td>0.460610</td>\n",
       "      <td>0.857309</td>\n",
       "      <td>0.277094</td>\n",
       "      <td>0.732757</td>\n",
       "      <td>0.070863</td>\n",
       "      <td>0.099313</td>\n",
       "      <td>0.895421</td>\n",
       "      <td>0.578508</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083012</td>\n",
       "      <td>0.805824</td>\n",
       "      <td>0.102142</td>\n",
       "      <td>0.043941</td>\n",
       "      <td>0.059123</td>\n",
       "      <td>0.036843</td>\n",
       "      <td>0.577638</td>\n",
       "      <td>0.034270</td>\n",
       "      <td>0.174444</td>\n",
       "      <td>0.045351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196432</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.916779</td>\n",
       "      <td>0.506125</td>\n",
       "      <td>0.823975</td>\n",
       "      <td>0.565643</td>\n",
       "      <td>0.848816</td>\n",
       "      <td>0.084068</td>\n",
       "      <td>0.232778</td>\n",
       "      <td>0.901168</td>\n",
       "      <td>0.628704</td>\n",
       "      <td>...</td>\n",
       "      <td>0.172733</td>\n",
       "      <td>0.926969</td>\n",
       "      <td>0.104711</td>\n",
       "      <td>0.047560</td>\n",
       "      <td>0.068287</td>\n",
       "      <td>0.043600</td>\n",
       "      <td>0.449071</td>\n",
       "      <td>0.032857</td>\n",
       "      <td>0.439872</td>\n",
       "      <td>0.053067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196506</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.939685</td>\n",
       "      <td>0.527137</td>\n",
       "      <td>0.712289</td>\n",
       "      <td>0.342275</td>\n",
       "      <td>0.810943</td>\n",
       "      <td>0.078483</td>\n",
       "      <td>0.089091</td>\n",
       "      <td>0.857550</td>\n",
       "      <td>0.629493</td>\n",
       "      <td>...</td>\n",
       "      <td>0.051485</td>\n",
       "      <td>0.727464</td>\n",
       "      <td>0.075486</td>\n",
       "      <td>0.031816</td>\n",
       "      <td>0.108818</td>\n",
       "      <td>0.031853</td>\n",
       "      <td>0.403560</td>\n",
       "      <td>0.024537</td>\n",
       "      <td>0.085534</td>\n",
       "      <td>0.053363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196507</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.815038</td>\n",
       "      <td>0.561500</td>\n",
       "      <td>0.735023</td>\n",
       "      <td>0.300820</td>\n",
       "      <td>0.830674</td>\n",
       "      <td>0.085613</td>\n",
       "      <td>0.076625</td>\n",
       "      <td>0.851060</td>\n",
       "      <td>0.626565</td>\n",
       "      <td>...</td>\n",
       "      <td>0.055969</td>\n",
       "      <td>0.875060</td>\n",
       "      <td>0.072526</td>\n",
       "      <td>0.044406</td>\n",
       "      <td>0.048262</td>\n",
       "      <td>0.032822</td>\n",
       "      <td>0.539247</td>\n",
       "      <td>0.036141</td>\n",
       "      <td>0.130871</td>\n",
       "      <td>0.039618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196508</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.561779</td>\n",
       "      <td>0.482367</td>\n",
       "      <td>0.824379</td>\n",
       "      <td>0.312059</td>\n",
       "      <td>0.865127</td>\n",
       "      <td>0.100088</td>\n",
       "      <td>0.102076</td>\n",
       "      <td>0.894283</td>\n",
       "      <td>0.600336</td>\n",
       "      <td>...</td>\n",
       "      <td>0.079980</td>\n",
       "      <td>0.787975</td>\n",
       "      <td>0.122041</td>\n",
       "      <td>0.045710</td>\n",
       "      <td>0.093470</td>\n",
       "      <td>0.043026</td>\n",
       "      <td>0.501775</td>\n",
       "      <td>0.028701</td>\n",
       "      <td>0.056252</td>\n",
       "      <td>0.035059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196509</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.883203</td>\n",
       "      <td>0.434855</td>\n",
       "      <td>0.797928</td>\n",
       "      <td>0.343013</td>\n",
       "      <td>0.758200</td>\n",
       "      <td>0.114223</td>\n",
       "      <td>0.184925</td>\n",
       "      <td>0.860588</td>\n",
       "      <td>0.568541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.097542</td>\n",
       "      <td>0.762363</td>\n",
       "      <td>0.158570</td>\n",
       "      <td>0.054024</td>\n",
       "      <td>0.108634</td>\n",
       "      <td>0.066233</td>\n",
       "      <td>0.458448</td>\n",
       "      <td>0.040045</td>\n",
       "      <td>0.089391</td>\n",
       "      <td>0.090369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1196510</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.890838</td>\n",
       "      <td>0.641983</td>\n",
       "      <td>0.724440</td>\n",
       "      <td>0.261995</td>\n",
       "      <td>0.620545</td>\n",
       "      <td>0.166151</td>\n",
       "      <td>0.138610</td>\n",
       "      <td>0.853633</td>\n",
       "      <td>0.554522</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128310</td>\n",
       "      <td>0.899229</td>\n",
       "      <td>0.212542</td>\n",
       "      <td>0.064727</td>\n",
       "      <td>0.166535</td>\n",
       "      <td>0.072853</td>\n",
       "      <td>0.518576</td>\n",
       "      <td>0.033771</td>\n",
       "      <td>0.095180</td>\n",
       "      <td>0.079189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows Ã— 11169 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Adrenocortical Cancer      A1BG       A2M     A2ML1    A4GALT  \\\n",
       "GSM1196428                    0.0  0.940233  0.749546  0.790944  0.374478   \n",
       "GSM1196429                    0.0  0.955381  0.245808  0.862709  0.458769   \n",
       "GSM1196430                    0.0  0.938320  0.694354  0.828572  0.362774   \n",
       "GSM1196431                    0.0  0.872892  0.460610  0.857309  0.277094   \n",
       "GSM1196432                    0.0  0.916779  0.506125  0.823975  0.565643   \n",
       "...                           ...       ...       ...       ...       ...   \n",
       "GSM1196506                    0.0  0.939685  0.527137  0.712289  0.342275   \n",
       "GSM1196507                    0.0  0.815038  0.561500  0.735023  0.300820   \n",
       "GSM1196508                    0.0  0.561779  0.482367  0.824379  0.312059   \n",
       "GSM1196509                    0.0  0.883203  0.434855  0.797928  0.343013   \n",
       "GSM1196510                    0.0  0.890838  0.641983  0.724440  0.261995   \n",
       "\n",
       "               A4GNT      AAAS      AACS     AADAC   AADACL2  ...    ZSCAN2  \\\n",
       "GSM1196428  0.634482  0.052668  0.217289  0.830051  0.672389  ...  0.040573   \n",
       "GSM1196429  0.861004  0.062860  0.070667  0.920872  0.661086  ...  0.054629   \n",
       "GSM1196430  0.768237  0.062411  0.085289  0.801509  0.542964  ...  0.040528   \n",
       "GSM1196431  0.732757  0.070863  0.099313  0.895421  0.578508  ...  0.083012   \n",
       "GSM1196432  0.848816  0.084068  0.232778  0.901168  0.628704  ...  0.172733   \n",
       "...              ...       ...       ...       ...       ...  ...       ...   \n",
       "GSM1196506  0.810943  0.078483  0.089091  0.857550  0.629493  ...  0.051485   \n",
       "GSM1196507  0.830674  0.085613  0.076625  0.851060  0.626565  ...  0.055969   \n",
       "GSM1196508  0.865127  0.100088  0.102076  0.894283  0.600336  ...  0.079980   \n",
       "GSM1196509  0.758200  0.114223  0.184925  0.860588  0.568541  ...  0.097542   \n",
       "GSM1196510  0.620545  0.166151  0.138610  0.853633  0.554522  ...  0.128310   \n",
       "\n",
       "              ZSCAN4    ZSWIM1      ZW10    ZWILCH     ZWINT      ZXDA  \\\n",
       "GSM1196428  0.810611  0.046088  0.035631  0.037018  0.031971  0.468787   \n",
       "GSM1196429  0.947878  0.069931  0.035145  0.036839  0.037855  0.468326   \n",
       "GSM1196430  0.850798  0.063188  0.038473  0.043276  0.033131  0.457744   \n",
       "GSM1196431  0.805824  0.102142  0.043941  0.059123  0.036843  0.577638   \n",
       "GSM1196432  0.926969  0.104711  0.047560  0.068287  0.043600  0.449071   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "GSM1196506  0.727464  0.075486  0.031816  0.108818  0.031853  0.403560   \n",
       "GSM1196507  0.875060  0.072526  0.044406  0.048262  0.032822  0.539247   \n",
       "GSM1196508  0.787975  0.122041  0.045710  0.093470  0.043026  0.501775   \n",
       "GSM1196509  0.762363  0.158570  0.054024  0.108634  0.066233  0.458448   \n",
       "GSM1196510  0.899229  0.212542  0.064727  0.166535  0.072853  0.518576   \n",
       "\n",
       "                 ZYX     ZZEF1      ZZZ3  \n",
       "GSM1196428  0.029696  0.079243  0.022379  \n",
       "GSM1196429  0.019655  0.072206  0.044275  \n",
       "GSM1196430  0.029563  0.066787  0.047593  \n",
       "GSM1196431  0.034270  0.174444  0.045351  \n",
       "GSM1196432  0.032857  0.439872  0.053067  \n",
       "...              ...       ...       ...  \n",
       "GSM1196506  0.024537  0.085534  0.053363  \n",
       "GSM1196507  0.036141  0.130871  0.039618  \n",
       "GSM1196508  0.028701  0.056252  0.035059  \n",
       "GSM1196509  0.040045  0.089391  0.090369  \n",
       "GSM1196510  0.033771  0.095180  0.079189  \n",
       "\n",
       "[81 rows x 11169 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 81 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '0.0' with 81 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"SNP data from 45 adrenocortical carcinomas\"\n",
      "!Series_summary\t\"SNP array data from 45 adrenocortical carcinomas were used to detect recurrent copy number alterations.\"\n",
      "!Series_overall_design\t\"7 tumors were analyzed with Illumina Human610-Quad v1.0 BeadChip. 38 tumors were analyzed with Illumina HumanOmniExpress BeadChip.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1196390</th>\n",
       "      <th>GSM1196391</th>\n",
       "      <th>GSM1196392</th>\n",
       "      <th>GSM1196393</th>\n",
       "      <th>GSM1196394</th>\n",
       "      <th>GSM1196395</th>\n",
       "      <th>GSM1196396</th>\n",
       "      <th>GSM1196397</th>\n",
       "      <th>GSM1196398</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196418</th>\n",
       "      <th>GSM1196419</th>\n",
       "      <th>GSM1196420</th>\n",
       "      <th>GSM1196421</th>\n",
       "      <th>GSM1196422</th>\n",
       "      <th>GSM1196423</th>\n",
       "      <th>GSM1196424</th>\n",
       "      <th>GSM1196425</th>\n",
       "      <th>GSM1196426</th>\n",
       "      <th>GSM1196427</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM1196390  \\\n",
       "0  !Sample_characteristics_ch1  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196391                           GSM1196392  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196393                           GSM1196394  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196395                           GSM1196396  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196397                           GSM1196398  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "   ...                           GSM1196418  \\\n",
       "0  ...  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196419                           GSM1196420  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196421                           GSM1196422  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196423                           GSM1196424  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196425                           GSM1196426  \\\n",
       "0  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196427  \n",
       "0  cell type: Adrenocortical carcinoma  \n",
       "\n",
       "[1 rows x 39 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: MemoryError\n",
    "\n",
    "cohort = accession_num = \"GSE49276\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'cell type: Adrenocortical carcinoma'], dtype=object)"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'cell type: Adrenocortical carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196390</th>\n",
       "      <th>GSM1196391</th>\n",
       "      <th>GSM1196392</th>\n",
       "      <th>GSM1196393</th>\n",
       "      <th>GSM1196394</th>\n",
       "      <th>GSM1196395</th>\n",
       "      <th>GSM1196396</th>\n",
       "      <th>GSM1196397</th>\n",
       "      <th>GSM1196398</th>\n",
       "      <th>GSM1196399</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196418</th>\n",
       "      <th>GSM1196419</th>\n",
       "      <th>GSM1196420</th>\n",
       "      <th>GSM1196421</th>\n",
       "      <th>GSM1196422</th>\n",
       "      <th>GSM1196423</th>\n",
       "      <th>GSM1196424</th>\n",
       "      <th>GSM1196425</th>\n",
       "      <th>GSM1196426</th>\n",
       "      <th>GSM1196427</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1196390  GSM1196391  GSM1196392  GSM1196393  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM1196394  GSM1196395  GSM1196396  GSM1196397  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM1196398  GSM1196399  ...  GSM1196418  GSM1196419  \\\n",
       "Adrenocortical Cancer           1           1  ...           1           1   \n",
       "\n",
       "                       GSM1196420  GSM1196421  GSM1196422  GSM1196423  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM1196424  GSM1196425  GSM1196426  GSM1196427  \n",
       "Adrenocortical Cancer           1           1           1           1  \n",
       "\n",
       "[1 rows x 38 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196390</th>\n",
       "      <th>GSM1196391</th>\n",
       "      <th>GSM1196392</th>\n",
       "      <th>GSM1196393</th>\n",
       "      <th>GSM1196394</th>\n",
       "      <th>GSM1196395</th>\n",
       "      <th>GSM1196396</th>\n",
       "      <th>GSM1196397</th>\n",
       "      <th>GSM1196398</th>\n",
       "      <th>GSM1196399</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196418</th>\n",
       "      <th>GSM1196419</th>\n",
       "      <th>GSM1196420</th>\n",
       "      <th>GSM1196421</th>\n",
       "      <th>GSM1196422</th>\n",
       "      <th>GSM1196423</th>\n",
       "      <th>GSM1196424</th>\n",
       "      <th>GSM1196425</th>\n",
       "      <th>GSM1196426</th>\n",
       "      <th>GSM1196427</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs1000000</th>\n",
       "      <td>0.1065</td>\n",
       "      <td>0.2021</td>\n",
       "      <td>-0.0258</td>\n",
       "      <td>0.0714</td>\n",
       "      <td>0.1821</td>\n",
       "      <td>0.3716</td>\n",
       "      <td>-0.0343</td>\n",
       "      <td>0.0632</td>\n",
       "      <td>0.4660</td>\n",
       "      <td>0.1609</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0428</td>\n",
       "      <td>-0.0564</td>\n",
       "      <td>0.0930</td>\n",
       "      <td>0.1396</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0.0441</td>\n",
       "      <td>0.0616</td>\n",
       "      <td>0.1482</td>\n",
       "      <td>0.0390</td>\n",
       "      <td>0.0873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000002</th>\n",
       "      <td>0.1184</td>\n",
       "      <td>-0.2697</td>\n",
       "      <td>-0.5366</td>\n",
       "      <td>-0.3392</td>\n",
       "      <td>0.1135</td>\n",
       "      <td>-0.5014</td>\n",
       "      <td>-0.2013</td>\n",
       "      <td>-0.3576</td>\n",
       "      <td>0.0717</td>\n",
       "      <td>-0.1686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1097</td>\n",
       "      <td>0.1616</td>\n",
       "      <td>-0.2828</td>\n",
       "      <td>-0.2522</td>\n",
       "      <td>0.2335</td>\n",
       "      <td>-0.1791</td>\n",
       "      <td>0.2011</td>\n",
       "      <td>0.2515</td>\n",
       "      <td>-0.5280</td>\n",
       "      <td>-0.1804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs10000023</th>\n",
       "      <td>-0.0627</td>\n",
       "      <td>0.0830</td>\n",
       "      <td>-0.1168</td>\n",
       "      <td>-0.0207</td>\n",
       "      <td>-0.0500</td>\n",
       "      <td>0.1033</td>\n",
       "      <td>-0.0169</td>\n",
       "      <td>0.0197</td>\n",
       "      <td>0.2101</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1189</td>\n",
       "      <td>0.0611</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>-0.0478</td>\n",
       "      <td>0.1226</td>\n",
       "      <td>-0.2908</td>\n",
       "      <td>0.1440</td>\n",
       "      <td>0.0704</td>\n",
       "      <td>-0.0951</td>\n",
       "      <td>-0.1550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000003</th>\n",
       "      <td>-0.0874</td>\n",
       "      <td>-0.5206</td>\n",
       "      <td>-0.2224</td>\n",
       "      <td>-0.1689</td>\n",
       "      <td>-0.0666</td>\n",
       "      <td>-0.2013</td>\n",
       "      <td>-0.1730</td>\n",
       "      <td>0.1010</td>\n",
       "      <td>0.2544</td>\n",
       "      <td>-0.1112</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0022</td>\n",
       "      <td>0.0930</td>\n",
       "      <td>-0.1605</td>\n",
       "      <td>-0.1667</td>\n",
       "      <td>0.0316</td>\n",
       "      <td>0.0493</td>\n",
       "      <td>-0.1910</td>\n",
       "      <td>0.0574</td>\n",
       "      <td>-0.1625</td>\n",
       "      <td>-0.1756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs10000030</th>\n",
       "      <td>0.7767</td>\n",
       "      <td>-0.0285</td>\n",
       "      <td>0.4749</td>\n",
       "      <td>0.2115</td>\n",
       "      <td>-0.0876</td>\n",
       "      <td>-0.3540</td>\n",
       "      <td>0.4442</td>\n",
       "      <td>-0.5963</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>-0.4139</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3487</td>\n",
       "      <td>0.7111</td>\n",
       "      <td>0.5774</td>\n",
       "      <td>0.4109</td>\n",
       "      <td>0.3663</td>\n",
       "      <td>0.2121</td>\n",
       "      <td>0.6078</td>\n",
       "      <td>0.8196</td>\n",
       "      <td>0.2831</td>\n",
       "      <td>0.8315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34742</th>\n",
       "      <td>0.1940</td>\n",
       "      <td>0.2780</td>\n",
       "      <td>0.2586</td>\n",
       "      <td>-0.1372</td>\n",
       "      <td>0.3958</td>\n",
       "      <td>0.1807</td>\n",
       "      <td>0.2403</td>\n",
       "      <td>0.1901</td>\n",
       "      <td>0.3416</td>\n",
       "      <td>0.4133</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2485</td>\n",
       "      <td>0.0686</td>\n",
       "      <td>0.1226</td>\n",
       "      <td>-0.0354</td>\n",
       "      <td>-0.1239</td>\n",
       "      <td>-0.3011</td>\n",
       "      <td>0.0032</td>\n",
       "      <td>-0.0218</td>\n",
       "      <td>-0.4481</td>\n",
       "      <td>0.0698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34743</th>\n",
       "      <td>0.2213</td>\n",
       "      <td>0.3447</td>\n",
       "      <td>0.2913</td>\n",
       "      <td>-0.2419</td>\n",
       "      <td>0.3323</td>\n",
       "      <td>0.1770</td>\n",
       "      <td>0.2181</td>\n",
       "      <td>0.0668</td>\n",
       "      <td>0.3055</td>\n",
       "      <td>0.4727</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2333</td>\n",
       "      <td>0.2030</td>\n",
       "      <td>0.1948</td>\n",
       "      <td>0.0367</td>\n",
       "      <td>0.0499</td>\n",
       "      <td>-0.1797</td>\n",
       "      <td>0.1574</td>\n",
       "      <td>-0.0151</td>\n",
       "      <td>-0.3437</td>\n",
       "      <td>0.1108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34744</th>\n",
       "      <td>-0.1274</td>\n",
       "      <td>-0.0201</td>\n",
       "      <td>-0.0016</td>\n",
       "      <td>-0.4076</td>\n",
       "      <td>0.2235</td>\n",
       "      <td>-0.0196</td>\n",
       "      <td>-0.0525</td>\n",
       "      <td>-0.0919</td>\n",
       "      <td>0.2553</td>\n",
       "      <td>0.2095</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0968</td>\n",
       "      <td>-0.0961</td>\n",
       "      <td>-0.1108</td>\n",
       "      <td>-0.4564</td>\n",
       "      <td>-0.4473</td>\n",
       "      <td>-0.7310</td>\n",
       "      <td>-0.1037</td>\n",
       "      <td>-0.5226</td>\n",
       "      <td>-0.7002</td>\n",
       "      <td>-0.2425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS34761</th>\n",
       "      <td>-0.1124</td>\n",
       "      <td>0.0917</td>\n",
       "      <td>0.0726</td>\n",
       "      <td>-0.3928</td>\n",
       "      <td>0.2356</td>\n",
       "      <td>0.0827</td>\n",
       "      <td>-0.0679</td>\n",
       "      <td>0.1435</td>\n",
       "      <td>0.3957</td>\n",
       "      <td>0.2965</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2568</td>\n",
       "      <td>0.0615</td>\n",
       "      <td>0.0917</td>\n",
       "      <td>-0.1877</td>\n",
       "      <td>-0.1994</td>\n",
       "      <td>-0.4965</td>\n",
       "      <td>-0.0968</td>\n",
       "      <td>-0.2921</td>\n",
       "      <td>-0.5193</td>\n",
       "      <td>-0.0352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VGXS35706</th>\n",
       "      <td>0.0560</td>\n",
       "      <td>0.2881</td>\n",
       "      <td>0.2553</td>\n",
       "      <td>-0.3665</td>\n",
       "      <td>0.3393</td>\n",
       "      <td>-0.2476</td>\n",
       "      <td>0.1102</td>\n",
       "      <td>-0.2108</td>\n",
       "      <td>0.3091</td>\n",
       "      <td>0.3059</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5293</td>\n",
       "      <td>-0.5513</td>\n",
       "      <td>-0.4255</td>\n",
       "      <td>-0.5476</td>\n",
       "      <td>-1.1190</td>\n",
       "      <td>-1.4563</td>\n",
       "      <td>-0.9340</td>\n",
       "      <td>-0.8940</td>\n",
       "      <td>-1.8872</td>\n",
       "      <td>-0.3105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>710272 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM1196390  GSM1196391  GSM1196392  GSM1196393  GSM1196394  \\\n",
       "ID                                                                       \n",
       "rs1000000       0.1065      0.2021     -0.0258      0.0714      0.1821   \n",
       "rs1000002       0.1184     -0.2697     -0.5366     -0.3392      0.1135   \n",
       "rs10000023     -0.0627      0.0830     -0.1168     -0.0207     -0.0500   \n",
       "rs1000003      -0.0874     -0.5206     -0.2224     -0.1689     -0.0666   \n",
       "rs10000030      0.7767     -0.0285      0.4749      0.2115     -0.0876   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "VGXS34742       0.1940      0.2780      0.2586     -0.1372      0.3958   \n",
       "VGXS34743       0.2213      0.3447      0.2913     -0.2419      0.3323   \n",
       "VGXS34744      -0.1274     -0.0201     -0.0016     -0.4076      0.2235   \n",
       "VGXS34761      -0.1124      0.0917      0.0726     -0.3928      0.2356   \n",
       "VGXS35706       0.0560      0.2881      0.2553     -0.3665      0.3393   \n",
       "\n",
       "            GSM1196395  GSM1196396  GSM1196397  GSM1196398  GSM1196399  ...  \\\n",
       "ID                                                                      ...   \n",
       "rs1000000       0.3716     -0.0343      0.0632      0.4660      0.1609  ...   \n",
       "rs1000002      -0.5014     -0.2013     -0.3576      0.0717     -0.1686  ...   \n",
       "rs10000023      0.1033     -0.0169      0.0197      0.2101      0.0587  ...   \n",
       "rs1000003      -0.2013     -0.1730      0.1010      0.2544     -0.1112  ...   \n",
       "rs10000030     -0.3540      0.4442     -0.5963      0.0009     -0.4139  ...   \n",
       "...                ...         ...         ...         ...         ...  ...   \n",
       "VGXS34742       0.1807      0.2403      0.1901      0.3416      0.4133  ...   \n",
       "VGXS34743       0.1770      0.2181      0.0668      0.3055      0.4727  ...   \n",
       "VGXS34744      -0.0196     -0.0525     -0.0919      0.2553      0.2095  ...   \n",
       "VGXS34761       0.0827     -0.0679      0.1435      0.3957      0.2965  ...   \n",
       "VGXS35706      -0.2476      0.1102     -0.2108      0.3091      0.3059  ...   \n",
       "\n",
       "            GSM1196418  GSM1196419  GSM1196420  GSM1196421  GSM1196422  \\\n",
       "ID                                                                       \n",
       "rs1000000       0.0428     -0.0564      0.0930      0.1396      0.0075   \n",
       "rs1000002      -0.1097      0.1616     -0.2828     -0.2522      0.2335   \n",
       "rs10000023     -0.1189      0.0611      0.0374     -0.0478      0.1226   \n",
       "rs1000003       0.0022      0.0930     -0.1605     -0.1667      0.0316   \n",
       "rs10000030      0.3487      0.7111      0.5774      0.4109      0.3663   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "VGXS34742       0.2485      0.0686      0.1226     -0.0354     -0.1239   \n",
       "VGXS34743       0.2333      0.2030      0.1948      0.0367      0.0499   \n",
       "VGXS34744       0.0968     -0.0961     -0.1108     -0.4564     -0.4473   \n",
       "VGXS34761       0.2568      0.0615      0.0917     -0.1877     -0.1994   \n",
       "VGXS35706      -0.5293     -0.5513     -0.4255     -0.5476     -1.1190   \n",
       "\n",
       "            GSM1196423  GSM1196424  GSM1196425  GSM1196426  GSM1196427  \n",
       "ID                                                                      \n",
       "rs1000000       0.0441      0.0616      0.1482      0.0390      0.0873  \n",
       "rs1000002      -0.1791      0.2011      0.2515     -0.5280     -0.1804  \n",
       "rs10000023     -0.2908      0.1440      0.0704     -0.0951     -0.1550  \n",
       "rs1000003       0.0493     -0.1910      0.0574     -0.1625     -0.1756  \n",
       "rs10000030      0.2121      0.6078      0.8196      0.2831      0.8315  \n",
       "...                ...         ...         ...         ...         ...  \n",
       "VGXS34742      -0.3011      0.0032     -0.0218     -0.4481      0.0698  \n",
       "VGXS34743      -0.1797      0.1574     -0.0151     -0.3437      0.1108  \n",
       "VGXS34744      -0.7310     -0.1037     -0.5226     -0.7002     -0.2425  \n",
       "VGXS34761      -0.4965     -0.0968     -0.2921     -0.5193     -0.0352  \n",
       "VGXS35706      -1.4563     -0.9340     -0.8940     -1.8872     -0.3105  \n",
       "\n",
       "[710272 rows x 38 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[151], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m requires_gene_mapping \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_gene_mapping:\n\u001b[1;32m----> 4\u001b[0m     gene_annotation \u001b[38;5;241m=\u001b[39m \u001b[43mget_gene_annotation\u001b[49m\u001b[43m(\u001b[49m\u001b[43msoft_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      5\u001b[0m     gene_annotation_summary \u001b[38;5;241m=\u001b[39m preview_df(gene_annotation)\n\u001b[0;32m      6\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene_annotation_summary)\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:128\u001b[0m, in \u001b[0;36mget_gene_annotation\u001b[1;34m(file_path, prefixes)\u001b[0m\n\u001b[0;32m    126\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_gene_annotation\u001b[39m(file_path, prefixes\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m^\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    127\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Extract from a SOFT file the gene annotation data\"\"\"\u001b[39;00m\n\u001b[1;32m--> 128\u001b[0m     gene_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_content_by_prefix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprefixes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munselect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfile\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    129\u001b[0m \u001b[43m                                             \u001b[49m\u001b[43mreturn_df_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    130\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m gene_metadata[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:106\u001b[0m, in \u001b[0;36mfilter_content_by_prefix\u001b[1;34m(source, prefixes_a, prefixes_b, unselect, source_type, return_df_a, return_df_b)\u001b[0m\n\u001b[0;32m    104\u001b[0m filtered_content_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(filtered_lines_a)\n\u001b[0;32m    105\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m return_df_a:\n\u001b[1;32m--> 106\u001b[0m     filtered_content_a \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfiltered_content_a\u001b[49m\u001b[43m)\u001b[49m, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, low_memory\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, on_bad_lines\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mskip\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m    107\u001b[0m filtered_content_b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    108\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filtered_lines_b:\n",
      "\u001b[1;31mMemoryError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"SNP data from 76 adrenocortical carcinomas\"\n",
      "!Series_summary\t\"SNP array data from 76 adrenocortical carcinomas were used to detect recurrent copy number alterations.\"\n",
      "!Series_overall_design\t\"76 tumors were analyzed with Illumina HumanCore-12v1 BeadChip.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1262431</th>\n",
       "      <th>GSM1262432</th>\n",
       "      <th>GSM1262433</th>\n",
       "      <th>GSM1262434</th>\n",
       "      <th>GSM1262435</th>\n",
       "      <th>GSM1262436</th>\n",
       "      <th>GSM1262437</th>\n",
       "      <th>GSM1262438</th>\n",
       "      <th>GSM1262439</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1262497</th>\n",
       "      <th>GSM1262498</th>\n",
       "      <th>GSM1262499</th>\n",
       "      <th>GSM1262500</th>\n",
       "      <th>GSM1262501</th>\n",
       "      <th>GSM1262502</th>\n",
       "      <th>GSM1262503</th>\n",
       "      <th>GSM1262504</th>\n",
       "      <th>GSM1262505</th>\n",
       "      <th>GSM1262506</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 68</td>\n",
       "      <td>age: 79</td>\n",
       "      <td>age: 45</td>\n",
       "      <td>age: 68</td>\n",
       "      <td>age: 31</td>\n",
       "      <td>age: 63</td>\n",
       "      <td>age: 47</td>\n",
       "      <td>age: 43</td>\n",
       "      <td>age: 76</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 24</td>\n",
       "      <td>age: 72</td>\n",
       "      <td>age: 37</td>\n",
       "      <td>age: 49</td>\n",
       "      <td>age: 63</td>\n",
       "      <td>age: 60</td>\n",
       "      <td>age: 53</td>\n",
       "      <td>age: 49</td>\n",
       "      <td>age: 65</td>\n",
       "      <td>age: 29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM1262431  \\\n",
       "0  !Sample_characteristics_ch1                               Sex: F   \n",
       "1  !Sample_characteristics_ch1                              age: 68   \n",
       "2  !Sample_characteristics_ch1  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262432                           GSM1262433  \\\n",
       "0                               Sex: F                               Sex: F   \n",
       "1                              age: 79                              age: 45   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262434                           GSM1262435  \\\n",
       "0                               Sex: F                               Sex: M   \n",
       "1                              age: 68                              age: 31   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262436                           GSM1262437  \\\n",
       "0                               Sex: M                               Sex: F   \n",
       "1                              age: 63                              age: 47   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262438                           GSM1262439  \\\n",
       "0                               Sex: F                               Sex: F   \n",
       "1                              age: 43                              age: 76   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "   ...                           GSM1262497  \\\n",
       "0  ...                               Sex: F   \n",
       "1  ...                              age: 24   \n",
       "2  ...  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262498                           GSM1262499  \\\n",
       "0                               Sex: F                               Sex: F   \n",
       "1                              age: 72                              age: 37   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262500                           GSM1262501  \\\n",
       "0                               Sex: F                               Sex: M   \n",
       "1                              age: 49                              age: 63   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262502                           GSM1262503  \\\n",
       "0                               Sex: M                               Sex: F   \n",
       "1                              age: 60                              age: 53   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262504                           GSM1262505  \\\n",
       "0                               Sex: F                               Sex: F   \n",
       "1                              age: 49                              age: 65   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1262506  \n",
       "0                               Sex: F  \n",
       "1                              age: 29  \n",
       "2  cell type: Adrenocortical carcinoma  \n",
       "\n",
       "[3 rows x 77 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: MemoryError\n",
    "\n",
    "cohort = accession_num = \"GSE52296\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'cell type: Adrenocortical carcinoma'], dtype=object)"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = 1\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'cell type: Adrenocortical carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1262431</th>\n",
       "      <th>GSM1262432</th>\n",
       "      <th>GSM1262433</th>\n",
       "      <th>GSM1262434</th>\n",
       "      <th>GSM1262435</th>\n",
       "      <th>GSM1262436</th>\n",
       "      <th>GSM1262437</th>\n",
       "      <th>GSM1262438</th>\n",
       "      <th>GSM1262439</th>\n",
       "      <th>GSM1262440</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1262497</th>\n",
       "      <th>GSM1262498</th>\n",
       "      <th>GSM1262499</th>\n",
       "      <th>GSM1262500</th>\n",
       "      <th>GSM1262501</th>\n",
       "      <th>GSM1262502</th>\n",
       "      <th>GSM1262503</th>\n",
       "      <th>GSM1262504</th>\n",
       "      <th>GSM1262505</th>\n",
       "      <th>GSM1262506</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>68</td>\n",
       "      <td>79</td>\n",
       "      <td>45</td>\n",
       "      <td>68</td>\n",
       "      <td>31</td>\n",
       "      <td>63</td>\n",
       "      <td>47</td>\n",
       "      <td>43</td>\n",
       "      <td>76</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>72</td>\n",
       "      <td>37</td>\n",
       "      <td>49</td>\n",
       "      <td>63</td>\n",
       "      <td>60</td>\n",
       "      <td>53</td>\n",
       "      <td>49</td>\n",
       "      <td>65</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1262431  GSM1262432  GSM1262433  GSM1262434  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Age                            68          79          45          68   \n",
       "Gender                          1           1           1           1   \n",
       "\n",
       "                       GSM1262435  GSM1262436  GSM1262437  GSM1262438  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Age                            31          63          47          43   \n",
       "Gender                          0           0           1           1   \n",
       "\n",
       "                       GSM1262439  GSM1262440  ...  GSM1262497  GSM1262498  \\\n",
       "Adrenocortical Cancer           1           1  ...           1           1   \n",
       "Age                            76          37  ...          24          72   \n",
       "Gender                          1           0  ...           1           1   \n",
       "\n",
       "                       GSM1262499  GSM1262500  GSM1262501  GSM1262502  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Age                            37          49          63          60   \n",
       "Gender                          1           1           0           0   \n",
       "\n",
       "                       GSM1262503  GSM1262504  GSM1262505  GSM1262506  \n",
       "Adrenocortical Cancer           1           1           1           1  \n",
       "Age                            53          49          65          29  \n",
       "Gender                          1           1           1           1  \n",
       "\n",
       "[3 rows x 76 columns]"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1262431</th>\n",
       "      <th>GSM1262432</th>\n",
       "      <th>GSM1262433</th>\n",
       "      <th>GSM1262434</th>\n",
       "      <th>GSM1262435</th>\n",
       "      <th>GSM1262436</th>\n",
       "      <th>GSM1262437</th>\n",
       "      <th>GSM1262438</th>\n",
       "      <th>GSM1262439</th>\n",
       "      <th>GSM1262440</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1262497</th>\n",
       "      <th>GSM1262498</th>\n",
       "      <th>GSM1262499</th>\n",
       "      <th>GSM1262500</th>\n",
       "      <th>GSM1262501</th>\n",
       "      <th>GSM1262502</th>\n",
       "      <th>GSM1262503</th>\n",
       "      <th>GSM1262504</th>\n",
       "      <th>GSM1262505</th>\n",
       "      <th>GSM1262506</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1KG_10_101715768</th>\n",
       "      <td>-0.0171</td>\n",
       "      <td>-0.0790</td>\n",
       "      <td>-0.0767</td>\n",
       "      <td>-0.0574</td>\n",
       "      <td>0.1267</td>\n",
       "      <td>0.0599</td>\n",
       "      <td>-0.0041</td>\n",
       "      <td>-0.0891</td>\n",
       "      <td>-0.0882</td>\n",
       "      <td>0.0274</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0390</td>\n",
       "      <td>-0.0518</td>\n",
       "      <td>-0.0336</td>\n",
       "      <td>-0.1126</td>\n",
       "      <td>-0.0117</td>\n",
       "      <td>-0.0319</td>\n",
       "      <td>-0.0696</td>\n",
       "      <td>0.0085</td>\n",
       "      <td>-0.0883</td>\n",
       "      <td>-0.0143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1KG_10_102265185</th>\n",
       "      <td>0.7398</td>\n",
       "      <td>0.2588</td>\n",
       "      <td>0.7608</td>\n",
       "      <td>-0.9820</td>\n",
       "      <td>-0.2913</td>\n",
       "      <td>0.7423</td>\n",
       "      <td>0.9168</td>\n",
       "      <td>0.6207</td>\n",
       "      <td>1.0650</td>\n",
       "      <td>1.2469</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8523</td>\n",
       "      <td>0.9312</td>\n",
       "      <td>0.8324</td>\n",
       "      <td>0.7108</td>\n",
       "      <td>0.5138</td>\n",
       "      <td>0.6199</td>\n",
       "      <td>0.5734</td>\n",
       "      <td>0.8608</td>\n",
       "      <td>0.6259</td>\n",
       "      <td>0.6200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1KG_10_102584498</th>\n",
       "      <td>0.1463</td>\n",
       "      <td>-0.0861</td>\n",
       "      <td>-0.0004</td>\n",
       "      <td>0.0137</td>\n",
       "      <td>0.3353</td>\n",
       "      <td>0.1679</td>\n",
       "      <td>0.0609</td>\n",
       "      <td>-0.0952</td>\n",
       "      <td>-0.0129</td>\n",
       "      <td>0.0825</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0055</td>\n",
       "      <td>-0.0705</td>\n",
       "      <td>-0.0535</td>\n",
       "      <td>-0.1776</td>\n",
       "      <td>-0.0815</td>\n",
       "      <td>-0.0657</td>\n",
       "      <td>-0.1601</td>\n",
       "      <td>-0.0177</td>\n",
       "      <td>-0.0909</td>\n",
       "      <td>0.0407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1KG_10_105796247</th>\n",
       "      <td>0.4250</td>\n",
       "      <td>0.1208</td>\n",
       "      <td>0.4771</td>\n",
       "      <td>0.3891</td>\n",
       "      <td>0.6188</td>\n",
       "      <td>0.4318</td>\n",
       "      <td>0.4521</td>\n",
       "      <td>0.3531</td>\n",
       "      <td>0.4691</td>\n",
       "      <td>0.5237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5056</td>\n",
       "      <td>0.4456</td>\n",
       "      <td>0.3760</td>\n",
       "      <td>0.2916</td>\n",
       "      <td>0.2201</td>\n",
       "      <td>0.2975</td>\n",
       "      <td>0.2365</td>\n",
       "      <td>0.3229</td>\n",
       "      <td>0.3548</td>\n",
       "      <td>0.5135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1KG_10_1066786</th>\n",
       "      <td>-0.0786</td>\n",
       "      <td>-0.2808</td>\n",
       "      <td>-0.4123</td>\n",
       "      <td>-0.0665</td>\n",
       "      <td>0.0945</td>\n",
       "      <td>-0.0363</td>\n",
       "      <td>0.0108</td>\n",
       "      <td>-0.2025</td>\n",
       "      <td>-0.1018</td>\n",
       "      <td>-0.0130</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1158</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>-0.2127</td>\n",
       "      <td>-0.2758</td>\n",
       "      <td>-0.1879</td>\n",
       "      <td>-0.0461</td>\n",
       "      <td>-0.2037</td>\n",
       "      <td>-0.1851</td>\n",
       "      <td>-0.2983</td>\n",
       "      <td>-0.1691</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM1262431  GSM1262432  GSM1262433  GSM1262434  GSM1262435  \\\n",
       "ID                                                                             \n",
       "1KG_10_101715768     -0.0171     -0.0790     -0.0767     -0.0574      0.1267   \n",
       "1KG_10_102265185      0.7398      0.2588      0.7608     -0.9820     -0.2913   \n",
       "1KG_10_102584498      0.1463     -0.0861     -0.0004      0.0137      0.3353   \n",
       "1KG_10_105796247      0.4250      0.1208      0.4771      0.3891      0.6188   \n",
       "1KG_10_1066786       -0.0786     -0.2808     -0.4123     -0.0665      0.0945   \n",
       "\n",
       "                  GSM1262436  GSM1262437  GSM1262438  GSM1262439  GSM1262440  \\\n",
       "ID                                                                             \n",
       "1KG_10_101715768      0.0599     -0.0041     -0.0891     -0.0882      0.0274   \n",
       "1KG_10_102265185      0.7423      0.9168      0.6207      1.0650      1.2469   \n",
       "1KG_10_102584498      0.1679      0.0609     -0.0952     -0.0129      0.0825   \n",
       "1KG_10_105796247      0.4318      0.4521      0.3531      0.4691      0.5237   \n",
       "1KG_10_1066786       -0.0363      0.0108     -0.2025     -0.1018     -0.0130   \n",
       "\n",
       "                  ...  GSM1262497  GSM1262498  GSM1262499  GSM1262500  \\\n",
       "ID                ...                                                   \n",
       "1KG_10_101715768  ...      0.0390     -0.0518     -0.0336     -0.1126   \n",
       "1KG_10_102265185  ...      0.8523      0.9312      0.8324      0.7108   \n",
       "1KG_10_102584498  ...     -0.0055     -0.0705     -0.0535     -0.1776   \n",
       "1KG_10_105796247  ...      0.5056      0.4456      0.3760      0.2916   \n",
       "1KG_10_1066786    ...     -0.1158     -0.1200     -0.2127     -0.2758   \n",
       "\n",
       "                  GSM1262501  GSM1262502  GSM1262503  GSM1262504  GSM1262505  \\\n",
       "ID                                                                             \n",
       "1KG_10_101715768     -0.0117     -0.0319     -0.0696      0.0085     -0.0883   \n",
       "1KG_10_102265185      0.5138      0.6199      0.5734      0.8608      0.6259   \n",
       "1KG_10_102584498     -0.0815     -0.0657     -0.1601     -0.0177     -0.0909   \n",
       "1KG_10_105796247      0.2201      0.2975      0.2365      0.3229      0.3548   \n",
       "1KG_10_1066786       -0.1879     -0.0461     -0.2037     -0.1851     -0.2983   \n",
       "\n",
       "                  GSM1262506  \n",
       "ID                            \n",
       "1KG_10_101715768     -0.0143  \n",
       "1KG_10_102265185      0.6200  \n",
       "1KG_10_102584498      0.0407  \n",
       "1KG_10_105796247      0.5135  \n",
       "1KG_10_1066786       -0.1691  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[177], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m requires_gene_mapping \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_gene_mapping:\n\u001b[1;32m----> 4\u001b[0m     gene_annotation \u001b[38;5;241m=\u001b[39m \u001b[43mget_gene_annotation\u001b[49m\u001b[43m(\u001b[49m\u001b[43msoft_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      5\u001b[0m     gene_annotation_summary \u001b[38;5;241m=\u001b[39m preview_df(gene_annotation)\n\u001b[0;32m      6\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene_annotation_summary)\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:128\u001b[0m, in \u001b[0;36mget_gene_annotation\u001b[1;34m(file_path, prefixes)\u001b[0m\n\u001b[0;32m    126\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_gene_annotation\u001b[39m(file_path, prefixes\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m^\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    127\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Extract from a SOFT file the gene annotation data\"\"\"\u001b[39;00m\n\u001b[1;32m--> 128\u001b[0m     gene_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_content_by_prefix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprefixes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munselect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfile\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    129\u001b[0m \u001b[43m                                             \u001b[49m\u001b[43mreturn_df_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    130\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m gene_metadata[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:106\u001b[0m, in \u001b[0;36mfilter_content_by_prefix\u001b[1;34m(source, prefixes_a, prefixes_b, unselect, source_type, return_df_a, return_df_b)\u001b[0m\n\u001b[0;32m    104\u001b[0m filtered_content_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(filtered_lines_a)\n\u001b[0;32m    105\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m return_df_a:\n\u001b[1;32m--> 106\u001b[0m     filtered_content_a \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfiltered_content_a\u001b[49m\u001b[43m)\u001b[49m, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, low_memory\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, on_bad_lines\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mskip\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m    107\u001b[0m filtered_content_b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    108\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filtered_lines_b:\n",
      "\u001b[1;31mMemoryError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Comprehensive molecular profiling of children with recurrent cancer II\"\n",
      "!Series_summary\t\"to explore possible treatment targets and reasons for agressive children cacners by comprehensive molecular profiling on several platforms\"\n",
      "!Series_summary\t\"to explore copy number aberrations related to cancers\"\n",
      "!Series_overall_design\t\"diagnostics of children meeting the oncologist with recurrent or agressive cancers where treatment options have been exhausted\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2889381</th>\n",
       "      <th>GSM2889382</th>\n",
       "      <th>GSM2889383</th>\n",
       "      <th>GSM2889384</th>\n",
       "      <th>GSM2889385</th>\n",
       "      <th>GSM2889386</th>\n",
       "      <th>GSM2889387</th>\n",
       "      <th>GSM2889388</th>\n",
       "      <th>GSM2889389</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2889414</th>\n",
       "      <th>GSM2889415</th>\n",
       "      <th>GSM2889416</th>\n",
       "      <th>GSM2889417</th>\n",
       "      <th>GSM2889418</th>\n",
       "      <th>GSM2889419</th>\n",
       "      <th>GSM2889420</th>\n",
       "      <th>GSM2889421</th>\n",
       "      <th>GSM2889422</th>\n",
       "      <th>GSM2889423</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>condition: Atypical meningioma</td>\n",
       "      <td>condition: Choroid plexus carcinoma / Malignan...</td>\n",
       "      <td>condition: Pilocytisc/pilomyxoid astrocytoma</td>\n",
       "      <td>condition: Pleomorphic xanthoastrocytoma</td>\n",
       "      <td>condition: Mesoblastisc nephroma</td>\n",
       "      <td>condition: Signetringcell carcinoma</td>\n",
       "      <td>condition: Ganglioglioma / Diffuse astrocytoma</td>\n",
       "      <td>condition: Chondrosarkoma</td>\n",
       "      <td>condition: Chordoma, dedefferentiated/anaplati...</td>\n",
       "      <td>...</td>\n",
       "      <td>condition: Anaplastic ependymoma</td>\n",
       "      <td>condition: Enchodroms</td>\n",
       "      <td>condition: Pineoblastoma</td>\n",
       "      <td>condition: Osteochondroma</td>\n",
       "      <td>condition: Malignant peripheral nerve sheeth t...</td>\n",
       "      <td>condition: Ewing sarcoma</td>\n",
       "      <td>condition: Adrenocortical carcinoma</td>\n",
       "      <td>condition: Anaplastic ependymoma</td>\n",
       "      <td>condition: Rhabdomyosarcoma</td>\n",
       "      <td>condition: Ependymom</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                      GSM2889381  \\\n",
       "0  !Sample_characteristics_ch1  condition: Atypical meningioma   \n",
       "\n",
       "                                          GSM2889382  \\\n",
       "0  condition: Choroid plexus carcinoma / Malignan...   \n",
       "\n",
       "                                     GSM2889383  \\\n",
       "0  condition: Pilocytisc/pilomyxoid astrocytoma   \n",
       "\n",
       "                                 GSM2889384                        GSM2889385  \\\n",
       "0  condition: Pleomorphic xanthoastrocytoma  condition: Mesoblastisc nephroma   \n",
       "\n",
       "                            GSM2889386  \\\n",
       "0  condition: Signetringcell carcinoma   \n",
       "\n",
       "                                       GSM2889387                 GSM2889388  \\\n",
       "0  condition: Ganglioglioma / Diffuse astrocytoma  condition: Chondrosarkoma   \n",
       "\n",
       "                                          GSM2889389  ...  \\\n",
       "0  condition: Chordoma, dedefferentiated/anaplati...  ...   \n",
       "\n",
       "                         GSM2889414             GSM2889415  \\\n",
       "0  condition: Anaplastic ependymoma  condition: Enchodroms   \n",
       "\n",
       "                 GSM2889416                 GSM2889417  \\\n",
       "0  condition: Pineoblastoma  condition: Osteochondroma   \n",
       "\n",
       "                                          GSM2889418  \\\n",
       "0  condition: Malignant peripheral nerve sheeth t...   \n",
       "\n",
       "                 GSM2889419                           GSM2889420  \\\n",
       "0  condition: Ewing sarcoma  condition: Adrenocortical carcinoma   \n",
       "\n",
       "                         GSM2889421                   GSM2889422  \\\n",
       "0  condition: Anaplastic ependymoma  condition: Rhabdomyosarcoma   \n",
       "\n",
       "             GSM2889423  \n",
       "0  condition: Ependymom  \n",
       "\n",
       "[1 rows x 44 columns]"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE108088\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'condition: Atypical meningioma',\n",
       "       'condition: Choroid plexus carcinoma / Malignant peripheral nerve sheeth tumor',\n",
       "       'condition: Pilocytisc/pilomyxoid astrocytoma',\n",
       "       'condition: Pleomorphic xanthoastrocytoma',\n",
       "       'condition: Mesoblastisc nephroma',\n",
       "       'condition: Signetringcell carcinoma',\n",
       "       'condition: Ganglioglioma / Diffuse astrocytoma',\n",
       "       'condition: Chondrosarkoma',\n",
       "       'condition: Chordoma, dedefferentiated/anaplatic type (INI1-loss)',\n",
       "       'condition: Hepatoblastoma',\n",
       "       'condition: Diffuse midline glioma H3K27M-mutated',\n",
       "       'condition: Anaplastisc ependymoma',\n",
       "       'condition: Juvenile xanthogranuloma',\n",
       "       'condition: Anaplastisc pleomorfic xanthoastrocytoma / Glioblastoma',\n",
       "       'condition: Alveolar rhabdomyosarcoma',\n",
       "       'condition: Precursor T-lymphoblastic lymphoma',\n",
       "       'condition: Glioblastoma',\n",
       "       'condition: Malignant peripheral nerve sheeth tumor',\n",
       "       'condition: Pilocytic astrocytoma', 'condition: Nephroblastoma',\n",
       "       'condition: Neuroblastoma', 'condition: Ganglioneuroblastoma',\n",
       "       'condition: Anaplastic ependymoma',\n",
       "       'condition: Gastrointestinal neuroectodermal tumour',\n",
       "       'condition: Atypical neurocytoma',\n",
       "       'condition: Chondroblastic osteosarcoma', 'condition: Enchodroms',\n",
       "       'condition: Pineoblastoma', 'condition: Osteochondroma',\n",
       "       'condition: Ewing sarcoma', 'condition: Adrenocortical carcinoma',\n",
       "       'condition: Rhabdomyosarcoma', 'condition: Ependymom'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'condition: Adrenocortical carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2889381</th>\n",
       "      <th>GSM2889382</th>\n",
       "      <th>GSM2889383</th>\n",
       "      <th>GSM2889384</th>\n",
       "      <th>GSM2889385</th>\n",
       "      <th>GSM2889386</th>\n",
       "      <th>GSM2889387</th>\n",
       "      <th>GSM2889388</th>\n",
       "      <th>GSM2889389</th>\n",
       "      <th>GSM2889390</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2889414</th>\n",
       "      <th>GSM2889415</th>\n",
       "      <th>GSM2889416</th>\n",
       "      <th>GSM2889417</th>\n",
       "      <th>GSM2889418</th>\n",
       "      <th>GSM2889419</th>\n",
       "      <th>GSM2889420</th>\n",
       "      <th>GSM2889421</th>\n",
       "      <th>GSM2889422</th>\n",
       "      <th>GSM2889423</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM2889381  GSM2889382  GSM2889383  GSM2889384  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM2889385  GSM2889386  GSM2889387  GSM2889388  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM2889389  GSM2889390  ...  GSM2889414  GSM2889415  \\\n",
       "Adrenocortical Cancer           0           0  ...           0           0   \n",
       "\n",
       "                       GSM2889416  GSM2889417  GSM2889418  GSM2889419  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM2889420  GSM2889421  GSM2889422  GSM2889423  \n",
       "Adrenocortical Cancer           1           0           0           0  \n",
       "\n",
       "[1 rows x 43 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2889381</th>\n",
       "      <th>GSM2889382</th>\n",
       "      <th>GSM2889383</th>\n",
       "      <th>GSM2889384</th>\n",
       "      <th>GSM2889385</th>\n",
       "      <th>GSM2889386</th>\n",
       "      <th>GSM2889387</th>\n",
       "      <th>GSM2889388</th>\n",
       "      <th>GSM2889389</th>\n",
       "      <th>GSM2889390</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2889414</th>\n",
       "      <th>GSM2889415</th>\n",
       "      <th>GSM2889416</th>\n",
       "      <th>GSM2889417</th>\n",
       "      <th>GSM2889418</th>\n",
       "      <th>GSM2889419</th>\n",
       "      <th>GSM2889420</th>\n",
       "      <th>GSM2889421</th>\n",
       "      <th>GSM2889422</th>\n",
       "      <th>GSM2889423</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>11.632633</td>\n",
       "      <td>9.908213</td>\n",
       "      <td>11.827391</td>\n",
       "      <td>11.958660</td>\n",
       "      <td>8.923858</td>\n",
       "      <td>9.266323</td>\n",
       "      <td>11.089168</td>\n",
       "      <td>8.470315</td>\n",
       "      <td>10.613575</td>\n",
       "      <td>7.913814</td>\n",
       "      <td>...</td>\n",
       "      <td>11.752788</td>\n",
       "      <td>8.436344</td>\n",
       "      <td>9.346787</td>\n",
       "      <td>7.883638</td>\n",
       "      <td>9.192208</td>\n",
       "      <td>9.482101</td>\n",
       "      <td>10.579143</td>\n",
       "      <td>11.138810</td>\n",
       "      <td>10.251416</td>\n",
       "      <td>10.644898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>7.337010</td>\n",
       "      <td>10.201116</td>\n",
       "      <td>7.642338</td>\n",
       "      <td>7.346276</td>\n",
       "      <td>7.319823</td>\n",
       "      <td>7.954301</td>\n",
       "      <td>6.362022</td>\n",
       "      <td>6.081242</td>\n",
       "      <td>6.553467</td>\n",
       "      <td>7.600938</td>\n",
       "      <td>...</td>\n",
       "      <td>7.270631</td>\n",
       "      <td>6.396677</td>\n",
       "      <td>5.713165</td>\n",
       "      <td>6.357230</td>\n",
       "      <td>6.809926</td>\n",
       "      <td>7.412866</td>\n",
       "      <td>8.555444</td>\n",
       "      <td>8.216679</td>\n",
       "      <td>7.847031</td>\n",
       "      <td>7.769231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>6.155619</td>\n",
       "      <td>5.734094</td>\n",
       "      <td>6.963633</td>\n",
       "      <td>6.015577</td>\n",
       "      <td>5.207735</td>\n",
       "      <td>5.381928</td>\n",
       "      <td>5.498432</td>\n",
       "      <td>5.975086</td>\n",
       "      <td>6.425805</td>\n",
       "      <td>7.771491</td>\n",
       "      <td>...</td>\n",
       "      <td>4.953888</td>\n",
       "      <td>5.893032</td>\n",
       "      <td>5.168161</td>\n",
       "      <td>6.312296</td>\n",
       "      <td>6.172038</td>\n",
       "      <td>5.837645</td>\n",
       "      <td>5.060681</td>\n",
       "      <td>6.882564</td>\n",
       "      <td>5.804388</td>\n",
       "      <td>4.921321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>6.661803</td>\n",
       "      <td>7.240843</td>\n",
       "      <td>7.367418</td>\n",
       "      <td>7.191843</td>\n",
       "      <td>8.568599</td>\n",
       "      <td>7.485777</td>\n",
       "      <td>5.858277</td>\n",
       "      <td>7.495252</td>\n",
       "      <td>7.265906</td>\n",
       "      <td>6.939669</td>\n",
       "      <td>...</td>\n",
       "      <td>6.998448</td>\n",
       "      <td>7.804581</td>\n",
       "      <td>8.771655</td>\n",
       "      <td>7.507523</td>\n",
       "      <td>7.867807</td>\n",
       "      <td>7.402482</td>\n",
       "      <td>7.488434</td>\n",
       "      <td>7.705329</td>\n",
       "      <td>7.487017</td>\n",
       "      <td>7.048244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>8.660903</td>\n",
       "      <td>2.994038</td>\n",
       "      <td>4.527065</td>\n",
       "      <td>3.175358</td>\n",
       "      <td>3.188646</td>\n",
       "      <td>3.323390</td>\n",
       "      <td>3.846578</td>\n",
       "      <td>3.374182</td>\n",
       "      <td>3.540232</td>\n",
       "      <td>3.046619</td>\n",
       "      <td>...</td>\n",
       "      <td>3.346600</td>\n",
       "      <td>3.068174</td>\n",
       "      <td>4.218592</td>\n",
       "      <td>3.139722</td>\n",
       "      <td>3.774415</td>\n",
       "      <td>3.929085</td>\n",
       "      <td>3.140174</td>\n",
       "      <td>3.061480</td>\n",
       "      <td>4.168771</td>\n",
       "      <td>3.022849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>6.998247</td>\n",
       "      <td>4.717028</td>\n",
       "      <td>4.837961</td>\n",
       "      <td>4.430565</td>\n",
       "      <td>5.083317</td>\n",
       "      <td>4.684815</td>\n",
       "      <td>9.259063</td>\n",
       "      <td>8.409200</td>\n",
       "      <td>8.789234</td>\n",
       "      <td>11.294144</td>\n",
       "      <td>...</td>\n",
       "      <td>7.720076</td>\n",
       "      <td>9.150521</td>\n",
       "      <td>4.780660</td>\n",
       "      <td>9.496483</td>\n",
       "      <td>10.338204</td>\n",
       "      <td>7.174550</td>\n",
       "      <td>6.740168</td>\n",
       "      <td>7.458485</td>\n",
       "      <td>6.663228</td>\n",
       "      <td>6.122767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>7.905597</td>\n",
       "      <td>4.778656</td>\n",
       "      <td>5.421778</td>\n",
       "      <td>4.496537</td>\n",
       "      <td>5.612980</td>\n",
       "      <td>4.905915</td>\n",
       "      <td>8.985258</td>\n",
       "      <td>9.384753</td>\n",
       "      <td>9.636073</td>\n",
       "      <td>11.218547</td>\n",
       "      <td>...</td>\n",
       "      <td>8.528286</td>\n",
       "      <td>10.324751</td>\n",
       "      <td>3.472218</td>\n",
       "      <td>10.550219</td>\n",
       "      <td>11.102371</td>\n",
       "      <td>7.850730</td>\n",
       "      <td>7.744783</td>\n",
       "      <td>8.414700</td>\n",
       "      <td>7.837027</td>\n",
       "      <td>7.030191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>2.632044</td>\n",
       "      <td>2.440085</td>\n",
       "      <td>2.370605</td>\n",
       "      <td>2.445787</td>\n",
       "      <td>2.425345</td>\n",
       "      <td>2.484874</td>\n",
       "      <td>2.594820</td>\n",
       "      <td>2.440124</td>\n",
       "      <td>2.819379</td>\n",
       "      <td>4.289839</td>\n",
       "      <td>...</td>\n",
       "      <td>2.576324</td>\n",
       "      <td>2.350293</td>\n",
       "      <td>2.389836</td>\n",
       "      <td>2.560649</td>\n",
       "      <td>2.454420</td>\n",
       "      <td>2.345227</td>\n",
       "      <td>2.506558</td>\n",
       "      <td>2.623461</td>\n",
       "      <td>2.471355</td>\n",
       "      <td>2.454033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>2.962745</td>\n",
       "      <td>3.074502</td>\n",
       "      <td>3.207309</td>\n",
       "      <td>3.055263</td>\n",
       "      <td>3.188459</td>\n",
       "      <td>2.964358</td>\n",
       "      <td>3.261142</td>\n",
       "      <td>3.112587</td>\n",
       "      <td>3.413876</td>\n",
       "      <td>3.218851</td>\n",
       "      <td>...</td>\n",
       "      <td>3.090914</td>\n",
       "      <td>3.072767</td>\n",
       "      <td>3.290665</td>\n",
       "      <td>3.125538</td>\n",
       "      <td>3.250400</td>\n",
       "      <td>2.972254</td>\n",
       "      <td>2.988300</td>\n",
       "      <td>3.147530</td>\n",
       "      <td>3.095629</td>\n",
       "      <td>2.929060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>2.793536</td>\n",
       "      <td>2.709928</td>\n",
       "      <td>2.709732</td>\n",
       "      <td>2.768932</td>\n",
       "      <td>2.773231</td>\n",
       "      <td>2.750002</td>\n",
       "      <td>3.086967</td>\n",
       "      <td>2.739828</td>\n",
       "      <td>2.883589</td>\n",
       "      <td>4.735059</td>\n",
       "      <td>...</td>\n",
       "      <td>2.711462</td>\n",
       "      <td>2.643172</td>\n",
       "      <td>3.044196</td>\n",
       "      <td>2.812862</td>\n",
       "      <td>2.756202</td>\n",
       "      <td>2.602647</td>\n",
       "      <td>2.886312</td>\n",
       "      <td>2.831216</td>\n",
       "      <td>2.649939</td>\n",
       "      <td>2.526819</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM2889381  GSM2889382  GSM2889383  GSM2889384  GSM2889385  \\\n",
       "ID                                                                            \n",
       "1007_s_at         11.632633    9.908213   11.827391   11.958660    8.923858   \n",
       "1053_at            7.337010   10.201116    7.642338    7.346276    7.319823   \n",
       "117_at             6.155619    5.734094    6.963633    6.015577    5.207735   \n",
       "121_at             6.661803    7.240843    7.367418    7.191843    8.568599   \n",
       "1255_g_at          8.660903    2.994038    4.527065    3.175358    3.188646   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     6.998247    4.717028    4.837961    4.430565    5.083317   \n",
       "AFFX-ThrX-M_at     7.905597    4.778656    5.421778    4.496537    5.612980   \n",
       "AFFX-TrpnX-3_at    2.632044    2.440085    2.370605    2.445787    2.425345   \n",
       "AFFX-TrpnX-5_at    2.962745    3.074502    3.207309    3.055263    3.188459   \n",
       "AFFX-TrpnX-M_at    2.793536    2.709928    2.709732    2.768932    2.773231   \n",
       "\n",
       "                 GSM2889386  GSM2889387  GSM2889388  GSM2889389  GSM2889390  \\\n",
       "ID                                                                            \n",
       "1007_s_at          9.266323   11.089168    8.470315   10.613575    7.913814   \n",
       "1053_at            7.954301    6.362022    6.081242    6.553467    7.600938   \n",
       "117_at             5.381928    5.498432    5.975086    6.425805    7.771491   \n",
       "121_at             7.485777    5.858277    7.495252    7.265906    6.939669   \n",
       "1255_g_at          3.323390    3.846578    3.374182    3.540232    3.046619   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     4.684815    9.259063    8.409200    8.789234   11.294144   \n",
       "AFFX-ThrX-M_at     4.905915    8.985258    9.384753    9.636073   11.218547   \n",
       "AFFX-TrpnX-3_at    2.484874    2.594820    2.440124    2.819379    4.289839   \n",
       "AFFX-TrpnX-5_at    2.964358    3.261142    3.112587    3.413876    3.218851   \n",
       "AFFX-TrpnX-M_at    2.750002    3.086967    2.739828    2.883589    4.735059   \n",
       "\n",
       "                 ...  GSM2889414  GSM2889415  GSM2889416  GSM2889417  \\\n",
       "ID               ...                                                   \n",
       "1007_s_at        ...   11.752788    8.436344    9.346787    7.883638   \n",
       "1053_at          ...    7.270631    6.396677    5.713165    6.357230   \n",
       "117_at           ...    4.953888    5.893032    5.168161    6.312296   \n",
       "121_at           ...    6.998448    7.804581    8.771655    7.507523   \n",
       "1255_g_at        ...    3.346600    3.068174    4.218592    3.139722   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at   ...    7.720076    9.150521    4.780660    9.496483   \n",
       "AFFX-ThrX-M_at   ...    8.528286   10.324751    3.472218   10.550219   \n",
       "AFFX-TrpnX-3_at  ...    2.576324    2.350293    2.389836    2.560649   \n",
       "AFFX-TrpnX-5_at  ...    3.090914    3.072767    3.290665    3.125538   \n",
       "AFFX-TrpnX-M_at  ...    2.711462    2.643172    3.044196    2.812862   \n",
       "\n",
       "                 GSM2889418  GSM2889419  GSM2889420  GSM2889421  GSM2889422  \\\n",
       "ID                                                                            \n",
       "1007_s_at          9.192208    9.482101   10.579143   11.138810   10.251416   \n",
       "1053_at            6.809926    7.412866    8.555444    8.216679    7.847031   \n",
       "117_at             6.172038    5.837645    5.060681    6.882564    5.804388   \n",
       "121_at             7.867807    7.402482    7.488434    7.705329    7.487017   \n",
       "1255_g_at          3.774415    3.929085    3.140174    3.061480    4.168771   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at    10.338204    7.174550    6.740168    7.458485    6.663228   \n",
       "AFFX-ThrX-M_at    11.102371    7.850730    7.744783    8.414700    7.837027   \n",
       "AFFX-TrpnX-3_at    2.454420    2.345227    2.506558    2.623461    2.471355   \n",
       "AFFX-TrpnX-5_at    3.250400    2.972254    2.988300    3.147530    3.095629   \n",
       "AFFX-TrpnX-M_at    2.756202    2.602647    2.886312    2.831216    2.649939   \n",
       "\n",
       "                 GSM2889423  \n",
       "ID                           \n",
       "1007_s_at         10.644898  \n",
       "1053_at            7.769231  \n",
       "117_at             4.921321  \n",
       "121_at             7.048244  \n",
       "1255_g_at          3.022849  \n",
       "...                     ...  \n",
       "AFFX-ThrX-5_at     6.122767  \n",
       "AFFX-ThrX-M_at     7.030191  \n",
       "AFFX-TrpnX-3_at    2.454033  \n",
       "AFFX-TrpnX-5_at    2.929060  \n",
       "AFFX-TrpnX-M_at    2.526819  \n",
       "\n",
       "[54675 rows x 43 columns]"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405738</th>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>6.122767291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405739</th>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>7.030190594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405740</th>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>2.454033124</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405741</th>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>2.929060487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405742</th>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>2.526819137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2405743 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID       GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0              1007_s_at       U48705     NaN            Homo sapiens   \n",
       "1                1053_at       M87338     NaN            Homo sapiens   \n",
       "2                 117_at       X51757     NaN            Homo sapiens   \n",
       "3                 121_at       X69699     NaN            Homo sapiens   \n",
       "4              1255_g_at       L36861     NaN            Homo sapiens   \n",
       "...                  ...          ...     ...                     ...   \n",
       "2405738   AFFX-ThrX-5_at  6.122767291     NaN                     NaN   \n",
       "2405739   AFFX-ThrX-M_at  7.030190594     NaN                     NaN   \n",
       "2405740  AFFX-TrpnX-3_at  2.454033124     NaN                     NaN   \n",
       "2405741  AFFX-TrpnX-5_at  2.929060487     NaN                     NaN   \n",
       "2405742  AFFX-TrpnX-M_at  2.526819137     NaN                     NaN   \n",
       "\n",
       "        Annotation Date      Sequence Type                  Sequence Source  \\\n",
       "0           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "2           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3           Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "4           Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                 ...                ...                              ...   \n",
       "2405738             NaN                NaN                              NaN   \n",
       "2405739             NaN                NaN                              NaN   \n",
       "2405740             NaN                NaN                              NaN   \n",
       "2405741             NaN                NaN                              NaN   \n",
       "2405742             NaN                NaN                              NaN   \n",
       "\n",
       "                                        Target Description  \\\n",
       "0        U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1        M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2        X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3        X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4        L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                    ...   \n",
       "2405738                                                NaN   \n",
       "2405739                                                NaN   \n",
       "2405740                                                NaN   \n",
       "2405741                                                NaN   \n",
       "2405742                                                NaN   \n",
       "\n",
       "        Representative Public ID  \\\n",
       "0                         U48705   \n",
       "1                         M87338   \n",
       "2                         X51757   \n",
       "3                         X69699   \n",
       "4                         L36861   \n",
       "...                          ...   \n",
       "2405738                      NaN   \n",
       "2405739                      NaN   \n",
       "2405740                      NaN   \n",
       "2405741                      NaN   \n",
       "2405742                      NaN   \n",
       "\n",
       "                                                Gene Title       Gene Symbol  \\\n",
       "0        discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1              replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2                     heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                             paired box 8              PAX8   \n",
       "4                  guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "...                                                    ...               ...   \n",
       "2405738                                                NaN               NaN   \n",
       "2405739                                                NaN               NaN   \n",
       "2405740                                                NaN               NaN   \n",
       "2405741                                                NaN               NaN   \n",
       "2405742                                                NaN               NaN   \n",
       "\n",
       "            ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0        780 /// 100616237  NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1                     5982  NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2                     3310                                          NM_002155   \n",
       "3                     7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                     2978                         NM_000409 /// XM_006715073   \n",
       "...                    ...                                                ...   \n",
       "2405738                NaN                                                NaN   \n",
       "2405739                NaN                                                NaN   \n",
       "2405740                NaN                                                NaN   \n",
       "2405741                NaN                                                NaN   \n",
       "2405742                NaN                                                NaN   \n",
       "\n",
       "                          Gene Ontology Biological Process  \\\n",
       "0        0001558 // regulation of cell growth // inferr...   \n",
       "1        0000278 // mitotic cell cycle // traceable aut...   \n",
       "2        0000902 // cell morphogenesis // inferred from...   \n",
       "3        0001655 // urogenital system development // in...   \n",
       "4        0007165 // signal transduction // non-traceabl...   \n",
       "...                                                    ...   \n",
       "2405738                                                NaN   \n",
       "2405739                                                NaN   \n",
       "2405740                                                NaN   \n",
       "2405741                                                NaN   \n",
       "2405742                                                NaN   \n",
       "\n",
       "                          Gene Ontology Cellular Component  \\\n",
       "0        0005576 // extracellular region // inferred fr...   \n",
       "1        0005634 // nucleus // inferred from electronic...   \n",
       "2        0005737 // cytoplasm // inferred from direct a...   \n",
       "3        0005634 // nucleus // inferred from direct ass...   \n",
       "4        0001750 // photoreceptor outer segment // infe...   \n",
       "...                                                    ...   \n",
       "2405738                                                NaN   \n",
       "2405739                                                NaN   \n",
       "2405740                                                NaN   \n",
       "2405741                                                NaN   \n",
       "2405742                                                NaN   \n",
       "\n",
       "                          Gene Ontology Molecular Function  \n",
       "0        0000166 // nucleotide binding // inferred from...  \n",
       "1        0000166 // nucleotide binding // inferred from...  \n",
       "2        0000166 // nucleotide binding // inferred from...  \n",
       "3        0000979 // RNA polymerase II core promoter seq...  \n",
       "4        0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                    ...  \n",
       "2405738                                                NaN  \n",
       "2405739                                                NaN  \n",
       "2405740                                                NaN  \n",
       "2405741                                                NaN  \n",
       "2405742                                                NaN  \n",
       "\n",
       "[2405743 rows x 16 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2889381</th>\n",
       "      <th>GSM2889382</th>\n",
       "      <th>GSM2889383</th>\n",
       "      <th>GSM2889384</th>\n",
       "      <th>GSM2889385</th>\n",
       "      <th>GSM2889386</th>\n",
       "      <th>GSM2889387</th>\n",
       "      <th>GSM2889388</th>\n",
       "      <th>GSM2889389</th>\n",
       "      <th>GSM2889390</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2889414</th>\n",
       "      <th>GSM2889415</th>\n",
       "      <th>GSM2889416</th>\n",
       "      <th>GSM2889417</th>\n",
       "      <th>GSM2889418</th>\n",
       "      <th>GSM2889419</th>\n",
       "      <th>GSM2889420</th>\n",
       "      <th>GSM2889421</th>\n",
       "      <th>GSM2889422</th>\n",
       "      <th>GSM2889423</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>6.961528</td>\n",
       "      <td>4.614781</td>\n",
       "      <td>7.307627</td>\n",
       "      <td>6.800363</td>\n",
       "      <td>5.353925</td>\n",
       "      <td>7.766458</td>\n",
       "      <td>7.201556</td>\n",
       "      <td>3.322718</td>\n",
       "      <td>4.282999</td>\n",
       "      <td>9.343120</td>\n",
       "      <td>...</td>\n",
       "      <td>6.685381</td>\n",
       "      <td>6.216733</td>\n",
       "      <td>5.839022</td>\n",
       "      <td>6.832801</td>\n",
       "      <td>4.510018</td>\n",
       "      <td>4.907617</td>\n",
       "      <td>10.999279</td>\n",
       "      <td>4.885687</td>\n",
       "      <td>5.614270</td>\n",
       "      <td>6.050458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>7.857824</td>\n",
       "      <td>8.475545</td>\n",
       "      <td>8.122764</td>\n",
       "      <td>8.262756</td>\n",
       "      <td>6.239403</td>\n",
       "      <td>7.024476</td>\n",
       "      <td>5.726159</td>\n",
       "      <td>4.983406</td>\n",
       "      <td>4.794567</td>\n",
       "      <td>8.003902</td>\n",
       "      <td>...</td>\n",
       "      <td>7.986060</td>\n",
       "      <td>6.931884</td>\n",
       "      <td>6.373791</td>\n",
       "      <td>7.243752</td>\n",
       "      <td>4.903729</td>\n",
       "      <td>4.942779</td>\n",
       "      <td>6.825852</td>\n",
       "      <td>7.099728</td>\n",
       "      <td>6.772416</td>\n",
       "      <td>8.103741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>7.857824</td>\n",
       "      <td>8.475545</td>\n",
       "      <td>8.122764</td>\n",
       "      <td>8.262756</td>\n",
       "      <td>6.239403</td>\n",
       "      <td>7.024476</td>\n",
       "      <td>5.726159</td>\n",
       "      <td>4.983406</td>\n",
       "      <td>4.794567</td>\n",
       "      <td>8.003902</td>\n",
       "      <td>...</td>\n",
       "      <td>7.986060</td>\n",
       "      <td>6.931884</td>\n",
       "      <td>6.373791</td>\n",
       "      <td>7.243752</td>\n",
       "      <td>4.903729</td>\n",
       "      <td>4.942779</td>\n",
       "      <td>6.825852</td>\n",
       "      <td>7.099728</td>\n",
       "      <td>6.772416</td>\n",
       "      <td>8.103741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1P2</th>\n",
       "      <td>3.282839</td>\n",
       "      <td>3.384542</td>\n",
       "      <td>3.583015</td>\n",
       "      <td>3.693280</td>\n",
       "      <td>3.659119</td>\n",
       "      <td>3.297807</td>\n",
       "      <td>3.077137</td>\n",
       "      <td>3.715791</td>\n",
       "      <td>3.521275</td>\n",
       "      <td>3.872615</td>\n",
       "      <td>...</td>\n",
       "      <td>3.373987</td>\n",
       "      <td>4.005146</td>\n",
       "      <td>3.765076</td>\n",
       "      <td>3.627202</td>\n",
       "      <td>3.900364</td>\n",
       "      <td>3.796208</td>\n",
       "      <td>3.813467</td>\n",
       "      <td>3.416828</td>\n",
       "      <td>3.381835</td>\n",
       "      <td>3.444124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC078883.4</th>\n",
       "      <td>4.918088</td>\n",
       "      <td>5.089238</td>\n",
       "      <td>5.002341</td>\n",
       "      <td>5.127558</td>\n",
       "      <td>5.015107</td>\n",
       "      <td>5.063883</td>\n",
       "      <td>4.670841</td>\n",
       "      <td>5.151271</td>\n",
       "      <td>5.563431</td>\n",
       "      <td>6.110059</td>\n",
       "      <td>...</td>\n",
       "      <td>5.280227</td>\n",
       "      <td>5.371916</td>\n",
       "      <td>5.862932</td>\n",
       "      <td>5.396481</td>\n",
       "      <td>5.211429</td>\n",
       "      <td>5.196257</td>\n",
       "      <td>5.523391</td>\n",
       "      <td>5.867906</td>\n",
       "      <td>5.196955</td>\n",
       "      <td>5.182184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abParts</th>\n",
       "      <td>10.281154</td>\n",
       "      <td>4.546864</td>\n",
       "      <td>4.466047</td>\n",
       "      <td>4.828189</td>\n",
       "      <td>6.367943</td>\n",
       "      <td>10.036273</td>\n",
       "      <td>5.342937</td>\n",
       "      <td>4.842982</td>\n",
       "      <td>5.424726</td>\n",
       "      <td>6.899275</td>\n",
       "      <td>...</td>\n",
       "      <td>4.452045</td>\n",
       "      <td>6.235608</td>\n",
       "      <td>5.254296</td>\n",
       "      <td>8.209720</td>\n",
       "      <td>4.999726</td>\n",
       "      <td>4.764322</td>\n",
       "      <td>4.767630</td>\n",
       "      <td>4.665493</td>\n",
       "      <td>10.966828</td>\n",
       "      <td>4.681283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>4.390690</td>\n",
       "      <td>3.319960</td>\n",
       "      <td>3.306106</td>\n",
       "      <td>3.474087</td>\n",
       "      <td>3.662178</td>\n",
       "      <td>5.427700</td>\n",
       "      <td>3.728478</td>\n",
       "      <td>3.518694</td>\n",
       "      <td>2.988476</td>\n",
       "      <td>5.461940</td>\n",
       "      <td>...</td>\n",
       "      <td>3.116597</td>\n",
       "      <td>3.572145</td>\n",
       "      <td>3.474087</td>\n",
       "      <td>5.356280</td>\n",
       "      <td>3.226779</td>\n",
       "      <td>3.634203</td>\n",
       "      <td>3.457508</td>\n",
       "      <td>3.818807</td>\n",
       "      <td>5.762159</td>\n",
       "      <td>3.293067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>av27s1</th>\n",
       "      <td>2.995613</td>\n",
       "      <td>3.093084</td>\n",
       "      <td>2.893304</td>\n",
       "      <td>3.044181</td>\n",
       "      <td>3.068038</td>\n",
       "      <td>4.723471</td>\n",
       "      <td>3.093084</td>\n",
       "      <td>3.435453</td>\n",
       "      <td>3.606269</td>\n",
       "      <td>4.498578</td>\n",
       "      <td>...</td>\n",
       "      <td>2.989656</td>\n",
       "      <td>3.916891</td>\n",
       "      <td>4.056041</td>\n",
       "      <td>3.388610</td>\n",
       "      <td>3.430777</td>\n",
       "      <td>3.056446</td>\n",
       "      <td>3.026054</td>\n",
       "      <td>3.031052</td>\n",
       "      <td>2.940405</td>\n",
       "      <td>2.991152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>4.176371</td>\n",
       "      <td>4.526591</td>\n",
       "      <td>4.814560</td>\n",
       "      <td>4.763223</td>\n",
       "      <td>4.640421</td>\n",
       "      <td>4.582546</td>\n",
       "      <td>8.159591</td>\n",
       "      <td>5.007854</td>\n",
       "      <td>4.850381</td>\n",
       "      <td>7.443819</td>\n",
       "      <td>...</td>\n",
       "      <td>4.318515</td>\n",
       "      <td>5.343503</td>\n",
       "      <td>6.769091</td>\n",
       "      <td>4.882332</td>\n",
       "      <td>4.976009</td>\n",
       "      <td>4.807871</td>\n",
       "      <td>4.682796</td>\n",
       "      <td>4.735828</td>\n",
       "      <td>4.525317</td>\n",
       "      <td>4.347550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mir-223</th>\n",
       "      <td>4.910865</td>\n",
       "      <td>4.090659</td>\n",
       "      <td>5.453438</td>\n",
       "      <td>5.108571</td>\n",
       "      <td>4.549454</td>\n",
       "      <td>6.372390</td>\n",
       "      <td>4.673962</td>\n",
       "      <td>4.376695</td>\n",
       "      <td>5.317572</td>\n",
       "      <td>4.973414</td>\n",
       "      <td>...</td>\n",
       "      <td>4.107590</td>\n",
       "      <td>7.066738</td>\n",
       "      <td>5.609446</td>\n",
       "      <td>5.401394</td>\n",
       "      <td>4.053826</td>\n",
       "      <td>4.311285</td>\n",
       "      <td>4.123832</td>\n",
       "      <td>5.140631</td>\n",
       "      <td>5.326176</td>\n",
       "      <td>4.018193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25851 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GSM2889381  GSM2889382  GSM2889383  GSM2889384  GSM2889385  \\\n",
       "Gene                                                                        \n",
       " ABCB4           6.961528    4.614781    7.307627    6.800363    5.353925   \n",
       " ABCC6P1         7.857824    8.475545    8.122764    8.262756    6.239403   \n",
       " ABCC6P2         7.857824    8.475545    8.122764    8.262756    6.239403   \n",
       " ABCD1P2         3.282839    3.384542    3.583015    3.693280    3.659119   \n",
       " AC078883.4      4.918088    5.089238    5.002341    5.127558    5.015107   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts         10.281154    4.546864    4.466047    4.828189    6.367943   \n",
       "alpha            4.390690    3.319960    3.306106    3.474087    3.662178   \n",
       "av27s1           2.995613    3.093084    2.893304    3.044181    3.068038   \n",
       "hsa-let-7a-3     4.176371    4.526591    4.814560    4.763223    4.640421   \n",
       "mir-223          4.910865    4.090659    5.453438    5.108571    4.549454   \n",
       "\n",
       "               GSM2889386  GSM2889387  GSM2889388  GSM2889389  GSM2889390  \\\n",
       "Gene                                                                        \n",
       " ABCB4           7.766458    7.201556    3.322718    4.282999    9.343120   \n",
       " ABCC6P1         7.024476    5.726159    4.983406    4.794567    8.003902   \n",
       " ABCC6P2         7.024476    5.726159    4.983406    4.794567    8.003902   \n",
       " ABCD1P2         3.297807    3.077137    3.715791    3.521275    3.872615   \n",
       " AC078883.4      5.063883    4.670841    5.151271    5.563431    6.110059   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts         10.036273    5.342937    4.842982    5.424726    6.899275   \n",
       "alpha            5.427700    3.728478    3.518694    2.988476    5.461940   \n",
       "av27s1           4.723471    3.093084    3.435453    3.606269    4.498578   \n",
       "hsa-let-7a-3     4.582546    8.159591    5.007854    4.850381    7.443819   \n",
       "mir-223          6.372390    4.673962    4.376695    5.317572    4.973414   \n",
       "\n",
       "               ...  GSM2889414  GSM2889415  GSM2889416  GSM2889417  \\\n",
       "Gene           ...                                                   \n",
       " ABCB4         ...    6.685381    6.216733    5.839022    6.832801   \n",
       " ABCC6P1       ...    7.986060    6.931884    6.373791    7.243752   \n",
       " ABCC6P2       ...    7.986060    6.931884    6.373791    7.243752   \n",
       " ABCD1P2       ...    3.373987    4.005146    3.765076    3.627202   \n",
       " AC078883.4    ...    5.280227    5.371916    5.862932    5.396481   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "abParts        ...    4.452045    6.235608    5.254296    8.209720   \n",
       "alpha          ...    3.116597    3.572145    3.474087    5.356280   \n",
       "av27s1         ...    2.989656    3.916891    4.056041    3.388610   \n",
       "hsa-let-7a-3   ...    4.318515    5.343503    6.769091    4.882332   \n",
       "mir-223        ...    4.107590    7.066738    5.609446    5.401394   \n",
       "\n",
       "               GSM2889418  GSM2889419  GSM2889420  GSM2889421  GSM2889422  \\\n",
       "Gene                                                                        \n",
       " ABCB4           4.510018    4.907617   10.999279    4.885687    5.614270   \n",
       " ABCC6P1         4.903729    4.942779    6.825852    7.099728    6.772416   \n",
       " ABCC6P2         4.903729    4.942779    6.825852    7.099728    6.772416   \n",
       " ABCD1P2         3.900364    3.796208    3.813467    3.416828    3.381835   \n",
       " AC078883.4      5.211429    5.196257    5.523391    5.867906    5.196955   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts          4.999726    4.764322    4.767630    4.665493   10.966828   \n",
       "alpha            3.226779    3.634203    3.457508    3.818807    5.762159   \n",
       "av27s1           3.430777    3.056446    3.026054    3.031052    2.940405   \n",
       "hsa-let-7a-3     4.976009    4.807871    4.682796    4.735828    4.525317   \n",
       "mir-223          4.053826    4.311285    4.123832    5.140631    5.326176   \n",
       "\n",
       "               GSM2889423  \n",
       "Gene                       \n",
       " ABCB4           6.050458  \n",
       " ABCC6P1         8.103741  \n",
       " ABCC6P2         8.103741  \n",
       " ABCD1P2         3.444124  \n",
       " AC078883.4      5.182184  \n",
       "...                   ...  \n",
       "abParts          4.681283  \n",
       "alpha            3.293067  \n",
       "av27s1           2.991152  \n",
       "hsa-let-7a-3     4.347550  \n",
       "mir-223          4.018193  \n",
       "\n",
       "[25851 rows x 43 columns]"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2889381</th>\n",
       "      <th>GSM2889382</th>\n",
       "      <th>GSM2889383</th>\n",
       "      <th>GSM2889384</th>\n",
       "      <th>GSM2889385</th>\n",
       "      <th>GSM2889386</th>\n",
       "      <th>GSM2889387</th>\n",
       "      <th>GSM2889388</th>\n",
       "      <th>GSM2889389</th>\n",
       "      <th>GSM2889390</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2889414</th>\n",
       "      <th>GSM2889415</th>\n",
       "      <th>GSM2889416</th>\n",
       "      <th>GSM2889417</th>\n",
       "      <th>GSM2889418</th>\n",
       "      <th>GSM2889419</th>\n",
       "      <th>GSM2889420</th>\n",
       "      <th>GSM2889421</th>\n",
       "      <th>GSM2889422</th>\n",
       "      <th>GSM2889423</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.523549</td>\n",
       "      <td>6.286303</td>\n",
       "      <td>4.722706</td>\n",
       "      <td>6.001011</td>\n",
       "      <td>4.748103</td>\n",
       "      <td>5.784339</td>\n",
       "      <td>5.073090</td>\n",
       "      <td>5.322978</td>\n",
       "      <td>5.552850</td>\n",
       "      <td>10.028951</td>\n",
       "      <td>...</td>\n",
       "      <td>6.278403</td>\n",
       "      <td>6.272556</td>\n",
       "      <td>4.600967</td>\n",
       "      <td>4.657533</td>\n",
       "      <td>5.614995</td>\n",
       "      <td>6.211944</td>\n",
       "      <td>5.638128</td>\n",
       "      <td>6.302610</td>\n",
       "      <td>5.382192</td>\n",
       "      <td>5.806563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>5.312258</td>\n",
       "      <td>5.636363</td>\n",
       "      <td>4.377456</td>\n",
       "      <td>4.989505</td>\n",
       "      <td>4.427621</td>\n",
       "      <td>4.967411</td>\n",
       "      <td>6.596631</td>\n",
       "      <td>4.705341</td>\n",
       "      <td>4.475880</td>\n",
       "      <td>6.932284</td>\n",
       "      <td>...</td>\n",
       "      <td>5.400511</td>\n",
       "      <td>5.276987</td>\n",
       "      <td>5.441775</td>\n",
       "      <td>4.602252</td>\n",
       "      <td>5.203488</td>\n",
       "      <td>4.949106</td>\n",
       "      <td>4.697487</td>\n",
       "      <td>4.967411</td>\n",
       "      <td>4.467902</td>\n",
       "      <td>4.628504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>3.603395</td>\n",
       "      <td>3.772548</td>\n",
       "      <td>3.782952</td>\n",
       "      <td>3.666064</td>\n",
       "      <td>4.272122</td>\n",
       "      <td>4.168318</td>\n",
       "      <td>3.925859</td>\n",
       "      <td>4.190043</td>\n",
       "      <td>4.335184</td>\n",
       "      <td>9.638197</td>\n",
       "      <td>...</td>\n",
       "      <td>3.803316</td>\n",
       "      <td>4.356464</td>\n",
       "      <td>5.655094</td>\n",
       "      <td>3.978726</td>\n",
       "      <td>4.182107</td>\n",
       "      <td>3.903059</td>\n",
       "      <td>3.856601</td>\n",
       "      <td>3.738244</td>\n",
       "      <td>3.695734</td>\n",
       "      <td>3.464957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>8.439460</td>\n",
       "      <td>6.234144</td>\n",
       "      <td>8.099240</td>\n",
       "      <td>7.572022</td>\n",
       "      <td>8.578123</td>\n",
       "      <td>7.410907</td>\n",
       "      <td>7.690870</td>\n",
       "      <td>7.885234</td>\n",
       "      <td>7.469755</td>\n",
       "      <td>9.430534</td>\n",
       "      <td>...</td>\n",
       "      <td>7.195041</td>\n",
       "      <td>7.508384</td>\n",
       "      <td>6.215682</td>\n",
       "      <td>8.150064</td>\n",
       "      <td>8.086973</td>\n",
       "      <td>7.541484</td>\n",
       "      <td>7.209844</td>\n",
       "      <td>7.469303</td>\n",
       "      <td>7.554861</td>\n",
       "      <td>6.911613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>6.784101</td>\n",
       "      <td>4.090659</td>\n",
       "      <td>5.154216</td>\n",
       "      <td>7.076977</td>\n",
       "      <td>6.310073</td>\n",
       "      <td>5.152857</td>\n",
       "      <td>5.589711</td>\n",
       "      <td>4.979747</td>\n",
       "      <td>4.542321</td>\n",
       "      <td>5.548948</td>\n",
       "      <td>...</td>\n",
       "      <td>5.360263</td>\n",
       "      <td>6.652004</td>\n",
       "      <td>6.177088</td>\n",
       "      <td>7.062321</td>\n",
       "      <td>6.452616</td>\n",
       "      <td>5.194912</td>\n",
       "      <td>5.040409</td>\n",
       "      <td>4.631870</td>\n",
       "      <td>4.431435</td>\n",
       "      <td>6.644617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>3.875443</td>\n",
       "      <td>6.901420</td>\n",
       "      <td>3.827137</td>\n",
       "      <td>3.950170</td>\n",
       "      <td>3.911654</td>\n",
       "      <td>3.895866</td>\n",
       "      <td>2.871190</td>\n",
       "      <td>4.332635</td>\n",
       "      <td>3.780255</td>\n",
       "      <td>4.327800</td>\n",
       "      <td>...</td>\n",
       "      <td>4.271786</td>\n",
       "      <td>3.688504</td>\n",
       "      <td>5.309030</td>\n",
       "      <td>4.069993</td>\n",
       "      <td>4.565107</td>\n",
       "      <td>4.002745</td>\n",
       "      <td>4.078990</td>\n",
       "      <td>3.747381</td>\n",
       "      <td>3.798240</td>\n",
       "      <td>4.433770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>8.324267</td>\n",
       "      <td>8.563336</td>\n",
       "      <td>8.407677</td>\n",
       "      <td>8.419080</td>\n",
       "      <td>8.368285</td>\n",
       "      <td>7.756354</td>\n",
       "      <td>7.599136</td>\n",
       "      <td>7.606243</td>\n",
       "      <td>7.776769</td>\n",
       "      <td>4.288961</td>\n",
       "      <td>...</td>\n",
       "      <td>7.871825</td>\n",
       "      <td>9.108076</td>\n",
       "      <td>4.800296</td>\n",
       "      <td>7.931003</td>\n",
       "      <td>7.282819</td>\n",
       "      <td>7.921015</td>\n",
       "      <td>8.153173</td>\n",
       "      <td>7.890696</td>\n",
       "      <td>7.555687</td>\n",
       "      <td>8.743453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>6.495728</td>\n",
       "      <td>7.109820</td>\n",
       "      <td>6.811351</td>\n",
       "      <td>7.635857</td>\n",
       "      <td>8.528983</td>\n",
       "      <td>7.296132</td>\n",
       "      <td>9.704778</td>\n",
       "      <td>7.226106</td>\n",
       "      <td>6.748325</td>\n",
       "      <td>9.603028</td>\n",
       "      <td>...</td>\n",
       "      <td>6.175848</td>\n",
       "      <td>6.383634</td>\n",
       "      <td>5.994175</td>\n",
       "      <td>7.025328</td>\n",
       "      <td>6.692828</td>\n",
       "      <td>6.590379</td>\n",
       "      <td>6.212945</td>\n",
       "      <td>6.775403</td>\n",
       "      <td>7.361366</td>\n",
       "      <td>5.754192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>5.451428</td>\n",
       "      <td>6.005562</td>\n",
       "      <td>5.749764</td>\n",
       "      <td>5.674700</td>\n",
       "      <td>5.761255</td>\n",
       "      <td>5.862683</td>\n",
       "      <td>5.487030</td>\n",
       "      <td>5.811785</td>\n",
       "      <td>5.731223</td>\n",
       "      <td>6.066631</td>\n",
       "      <td>...</td>\n",
       "      <td>5.459600</td>\n",
       "      <td>5.614055</td>\n",
       "      <td>6.096445</td>\n",
       "      <td>5.627089</td>\n",
       "      <td>5.651303</td>\n",
       "      <td>5.808518</td>\n",
       "      <td>6.284363</td>\n",
       "      <td>5.593404</td>\n",
       "      <td>5.630627</td>\n",
       "      <td>5.498746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>8.522046</td>\n",
       "      <td>8.770399</td>\n",
       "      <td>8.118409</td>\n",
       "      <td>8.082138</td>\n",
       "      <td>7.635416</td>\n",
       "      <td>8.030460</td>\n",
       "      <td>5.420636</td>\n",
       "      <td>7.711138</td>\n",
       "      <td>6.964051</td>\n",
       "      <td>4.419665</td>\n",
       "      <td>...</td>\n",
       "      <td>8.079711</td>\n",
       "      <td>7.378567</td>\n",
       "      <td>5.492084</td>\n",
       "      <td>7.758366</td>\n",
       "      <td>7.725530</td>\n",
       "      <td>7.964984</td>\n",
       "      <td>8.239172</td>\n",
       "      <td>8.386879</td>\n",
       "      <td>7.604333</td>\n",
       "      <td>8.283418</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM2889381  GSM2889382  GSM2889383  GSM2889384  GSM2889385  \\\n",
       "A1BG        5.523549    6.286303    4.722706    6.001011    4.748103   \n",
       "A1BG-AS1    5.312258    5.636363    4.377456    4.989505    4.427621   \n",
       "A1CF        3.603395    3.772548    3.782952    3.666064    4.272122   \n",
       "A2M         8.439460    6.234144    8.099240    7.572022    8.578123   \n",
       "A2M-AS1     6.784101    4.090659    5.154216    7.076977    6.310073   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A      3.875443    6.901420    3.827137    3.950170    3.911654   \n",
       "ZYG11B      8.324267    8.563336    8.407677    8.419080    8.368285   \n",
       "ZYX         6.495728    7.109820    6.811351    7.635857    8.528983   \n",
       "ZZEF1       5.451428    6.005562    5.749764    5.674700    5.761255   \n",
       "ZZZ3        8.522046    8.770399    8.118409    8.082138    7.635416   \n",
       "\n",
       "          GSM2889386  GSM2889387  GSM2889388  GSM2889389  GSM2889390  ...  \\\n",
       "A1BG        5.784339    5.073090    5.322978    5.552850   10.028951  ...   \n",
       "A1BG-AS1    4.967411    6.596631    4.705341    4.475880    6.932284  ...   \n",
       "A1CF        4.168318    3.925859    4.190043    4.335184    9.638197  ...   \n",
       "A2M         7.410907    7.690870    7.885234    7.469755    9.430534  ...   \n",
       "A2M-AS1     5.152857    5.589711    4.979747    4.542321    5.548948  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A      3.895866    2.871190    4.332635    3.780255    4.327800  ...   \n",
       "ZYG11B      7.756354    7.599136    7.606243    7.776769    4.288961  ...   \n",
       "ZYX         7.296132    9.704778    7.226106    6.748325    9.603028  ...   \n",
       "ZZEF1       5.862683    5.487030    5.811785    5.731223    6.066631  ...   \n",
       "ZZZ3        8.030460    5.420636    7.711138    6.964051    4.419665  ...   \n",
       "\n",
       "          GSM2889414  GSM2889415  GSM2889416  GSM2889417  GSM2889418  \\\n",
       "A1BG        6.278403    6.272556    4.600967    4.657533    5.614995   \n",
       "A1BG-AS1    5.400511    5.276987    5.441775    4.602252    5.203488   \n",
       "A1CF        3.803316    4.356464    5.655094    3.978726    4.182107   \n",
       "A2M         7.195041    7.508384    6.215682    8.150064    8.086973   \n",
       "A2M-AS1     5.360263    6.652004    6.177088    7.062321    6.452616   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A      4.271786    3.688504    5.309030    4.069993    4.565107   \n",
       "ZYG11B      7.871825    9.108076    4.800296    7.931003    7.282819   \n",
       "ZYX         6.175848    6.383634    5.994175    7.025328    6.692828   \n",
       "ZZEF1       5.459600    5.614055    6.096445    5.627089    5.651303   \n",
       "ZZZ3        8.079711    7.378567    5.492084    7.758366    7.725530   \n",
       "\n",
       "          GSM2889419  GSM2889420  GSM2889421  GSM2889422  GSM2889423  \n",
       "A1BG        6.211944    5.638128    6.302610    5.382192    5.806563  \n",
       "A1BG-AS1    4.949106    4.697487    4.967411    4.467902    4.628504  \n",
       "A1CF        3.903059    3.856601    3.738244    3.695734    3.464957  \n",
       "A2M         7.541484    7.209844    7.469303    7.554861    6.911613  \n",
       "A2M-AS1     5.194912    5.040409    4.631870    4.431435    6.644617  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "ZYG11A      4.002745    4.078990    3.747381    3.798240    4.433770  \n",
       "ZYG11B      7.921015    8.153173    7.890696    7.555687    8.743453  \n",
       "ZYX         6.590379    6.212945    6.775403    7.361366    5.754192  \n",
       "ZZEF1       5.808518    6.284363    5.593404    5.630627    5.498746  \n",
       "ZZZ3        7.964984    8.239172    8.386879    7.604333    8.283418  \n",
       "\n",
       "[17857 rows x 43 columns]"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2889381</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.523549</td>\n",
       "      <td>5.312258</td>\n",
       "      <td>3.603395</td>\n",
       "      <td>8.439460</td>\n",
       "      <td>6.784101</td>\n",
       "      <td>3.275863</td>\n",
       "      <td>3.821555</td>\n",
       "      <td>5.833153</td>\n",
       "      <td>3.929618</td>\n",
       "      <td>...</td>\n",
       "      <td>7.404137</td>\n",
       "      <td>8.066233</td>\n",
       "      <td>7.023525</td>\n",
       "      <td>5.423627</td>\n",
       "      <td>6.130173</td>\n",
       "      <td>3.875443</td>\n",
       "      <td>8.324267</td>\n",
       "      <td>6.495728</td>\n",
       "      <td>5.451428</td>\n",
       "      <td>8.522046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889382</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.286303</td>\n",
       "      <td>5.636363</td>\n",
       "      <td>3.772548</td>\n",
       "      <td>6.234144</td>\n",
       "      <td>4.090659</td>\n",
       "      <td>3.355778</td>\n",
       "      <td>3.929229</td>\n",
       "      <td>5.390436</td>\n",
       "      <td>4.152816</td>\n",
       "      <td>...</td>\n",
       "      <td>9.315457</td>\n",
       "      <td>11.280892</td>\n",
       "      <td>6.738129</td>\n",
       "      <td>5.336899</td>\n",
       "      <td>6.629826</td>\n",
       "      <td>6.901420</td>\n",
       "      <td>8.563336</td>\n",
       "      <td>7.109820</td>\n",
       "      <td>6.005562</td>\n",
       "      <td>8.770399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889383</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.722706</td>\n",
       "      <td>4.377456</td>\n",
       "      <td>3.782952</td>\n",
       "      <td>8.099240</td>\n",
       "      <td>5.154216</td>\n",
       "      <td>3.262292</td>\n",
       "      <td>4.774103</td>\n",
       "      <td>4.825479</td>\n",
       "      <td>4.381623</td>\n",
       "      <td>...</td>\n",
       "      <td>7.241154</td>\n",
       "      <td>7.623172</td>\n",
       "      <td>7.457557</td>\n",
       "      <td>5.477725</td>\n",
       "      <td>6.049950</td>\n",
       "      <td>3.827137</td>\n",
       "      <td>8.407677</td>\n",
       "      <td>6.811351</td>\n",
       "      <td>5.749764</td>\n",
       "      <td>8.118409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889384</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.001011</td>\n",
       "      <td>4.989505</td>\n",
       "      <td>3.666064</td>\n",
       "      <td>7.572022</td>\n",
       "      <td>7.076977</td>\n",
       "      <td>3.777149</td>\n",
       "      <td>4.317872</td>\n",
       "      <td>5.557962</td>\n",
       "      <td>4.262175</td>\n",
       "      <td>...</td>\n",
       "      <td>6.538005</td>\n",
       "      <td>6.921254</td>\n",
       "      <td>7.068052</td>\n",
       "      <td>5.379503</td>\n",
       "      <td>6.370581</td>\n",
       "      <td>3.950170</td>\n",
       "      <td>8.419080</td>\n",
       "      <td>7.635857</td>\n",
       "      <td>5.674700</td>\n",
       "      <td>8.082138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889385</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.748103</td>\n",
       "      <td>4.427621</td>\n",
       "      <td>4.272122</td>\n",
       "      <td>8.578123</td>\n",
       "      <td>6.310073</td>\n",
       "      <td>3.301956</td>\n",
       "      <td>4.272148</td>\n",
       "      <td>5.374137</td>\n",
       "      <td>4.317426</td>\n",
       "      <td>...</td>\n",
       "      <td>7.874055</td>\n",
       "      <td>7.974063</td>\n",
       "      <td>6.767503</td>\n",
       "      <td>4.873842</td>\n",
       "      <td>5.980389</td>\n",
       "      <td>3.911654</td>\n",
       "      <td>8.368285</td>\n",
       "      <td>8.528983</td>\n",
       "      <td>5.761255</td>\n",
       "      <td>7.635416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889386</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.784339</td>\n",
       "      <td>4.967411</td>\n",
       "      <td>4.168318</td>\n",
       "      <td>7.410907</td>\n",
       "      <td>5.152857</td>\n",
       "      <td>3.246321</td>\n",
       "      <td>3.977649</td>\n",
       "      <td>5.160725</td>\n",
       "      <td>6.856612</td>\n",
       "      <td>...</td>\n",
       "      <td>7.595532</td>\n",
       "      <td>9.056683</td>\n",
       "      <td>6.624460</td>\n",
       "      <td>5.334879</td>\n",
       "      <td>5.982906</td>\n",
       "      <td>3.895866</td>\n",
       "      <td>7.756354</td>\n",
       "      <td>7.296132</td>\n",
       "      <td>5.862683</td>\n",
       "      <td>8.030460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889387</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.073090</td>\n",
       "      <td>6.596631</td>\n",
       "      <td>3.925859</td>\n",
       "      <td>7.690870</td>\n",
       "      <td>5.589711</td>\n",
       "      <td>3.595074</td>\n",
       "      <td>3.510832</td>\n",
       "      <td>7.146452</td>\n",
       "      <td>3.627231</td>\n",
       "      <td>...</td>\n",
       "      <td>5.044171</td>\n",
       "      <td>2.995060</td>\n",
       "      <td>5.120191</td>\n",
       "      <td>6.015022</td>\n",
       "      <td>6.120873</td>\n",
       "      <td>2.871190</td>\n",
       "      <td>7.599136</td>\n",
       "      <td>9.704778</td>\n",
       "      <td>5.487030</td>\n",
       "      <td>5.420636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889388</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.322978</td>\n",
       "      <td>4.705341</td>\n",
       "      <td>4.190043</td>\n",
       "      <td>7.885234</td>\n",
       "      <td>4.979747</td>\n",
       "      <td>3.422809</td>\n",
       "      <td>3.838083</td>\n",
       "      <td>5.613246</td>\n",
       "      <td>4.328624</td>\n",
       "      <td>...</td>\n",
       "      <td>6.942579</td>\n",
       "      <td>7.837928</td>\n",
       "      <td>6.339747</td>\n",
       "      <td>4.788353</td>\n",
       "      <td>6.381847</td>\n",
       "      <td>4.332635</td>\n",
       "      <td>7.606243</td>\n",
       "      <td>7.226106</td>\n",
       "      <td>5.811785</td>\n",
       "      <td>7.711138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889389</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.552850</td>\n",
       "      <td>4.475880</td>\n",
       "      <td>4.335184</td>\n",
       "      <td>7.469755</td>\n",
       "      <td>4.542321</td>\n",
       "      <td>3.403456</td>\n",
       "      <td>4.072234</td>\n",
       "      <td>6.805424</td>\n",
       "      <td>5.046665</td>\n",
       "      <td>...</td>\n",
       "      <td>6.345089</td>\n",
       "      <td>8.777369</td>\n",
       "      <td>7.087475</td>\n",
       "      <td>5.236264</td>\n",
       "      <td>5.944496</td>\n",
       "      <td>3.780255</td>\n",
       "      <td>7.776769</td>\n",
       "      <td>6.748325</td>\n",
       "      <td>5.731223</td>\n",
       "      <td>6.964051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889390</th>\n",
       "      <td>0.0</td>\n",
       "      <td>10.028951</td>\n",
       "      <td>6.932284</td>\n",
       "      <td>9.638197</td>\n",
       "      <td>9.430534</td>\n",
       "      <td>5.548948</td>\n",
       "      <td>3.741756</td>\n",
       "      <td>4.228405</td>\n",
       "      <td>8.081186</td>\n",
       "      <td>4.556953</td>\n",
       "      <td>...</td>\n",
       "      <td>3.688915</td>\n",
       "      <td>2.822378</td>\n",
       "      <td>3.483630</td>\n",
       "      <td>5.673570</td>\n",
       "      <td>5.903220</td>\n",
       "      <td>4.327800</td>\n",
       "      <td>4.288961</td>\n",
       "      <td>9.603028</td>\n",
       "      <td>6.066631</td>\n",
       "      <td>4.419665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889391</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.417395</td>\n",
       "      <td>5.015280</td>\n",
       "      <td>3.714774</td>\n",
       "      <td>8.120436</td>\n",
       "      <td>4.947901</td>\n",
       "      <td>3.418168</td>\n",
       "      <td>4.211228</td>\n",
       "      <td>5.074946</td>\n",
       "      <td>4.096684</td>\n",
       "      <td>...</td>\n",
       "      <td>8.363608</td>\n",
       "      <td>9.731977</td>\n",
       "      <td>6.729941</td>\n",
       "      <td>5.507374</td>\n",
       "      <td>6.100880</td>\n",
       "      <td>4.129861</td>\n",
       "      <td>9.114931</td>\n",
       "      <td>7.631545</td>\n",
       "      <td>5.577168</td>\n",
       "      <td>8.143149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889392</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.975241</td>\n",
       "      <td>4.258458</td>\n",
       "      <td>3.563695</td>\n",
       "      <td>7.553463</td>\n",
       "      <td>4.995097</td>\n",
       "      <td>3.347005</td>\n",
       "      <td>3.990605</td>\n",
       "      <td>4.632107</td>\n",
       "      <td>4.108716</td>\n",
       "      <td>...</td>\n",
       "      <td>7.753695</td>\n",
       "      <td>10.239855</td>\n",
       "      <td>7.474382</td>\n",
       "      <td>5.820885</td>\n",
       "      <td>6.783218</td>\n",
       "      <td>3.810572</td>\n",
       "      <td>8.988548</td>\n",
       "      <td>6.962456</td>\n",
       "      <td>5.607036</td>\n",
       "      <td>7.965766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889393</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.219072</td>\n",
       "      <td>4.967411</td>\n",
       "      <td>3.835905</td>\n",
       "      <td>8.228831</td>\n",
       "      <td>5.810245</td>\n",
       "      <td>3.611951</td>\n",
       "      <td>4.472261</td>\n",
       "      <td>5.971737</td>\n",
       "      <td>4.770631</td>\n",
       "      <td>...</td>\n",
       "      <td>6.045949</td>\n",
       "      <td>7.027758</td>\n",
       "      <td>5.806202</td>\n",
       "      <td>4.721118</td>\n",
       "      <td>6.053315</td>\n",
       "      <td>4.553429</td>\n",
       "      <td>8.015609</td>\n",
       "      <td>7.105865</td>\n",
       "      <td>5.686411</td>\n",
       "      <td>7.388064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889394</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.911014</td>\n",
       "      <td>4.760648</td>\n",
       "      <td>3.614136</td>\n",
       "      <td>8.089677</td>\n",
       "      <td>4.868123</td>\n",
       "      <td>3.388084</td>\n",
       "      <td>3.817156</td>\n",
       "      <td>5.588154</td>\n",
       "      <td>4.077890</td>\n",
       "      <td>...</td>\n",
       "      <td>6.337353</td>\n",
       "      <td>8.070385</td>\n",
       "      <td>7.853956</td>\n",
       "      <td>5.449738</td>\n",
       "      <td>6.900473</td>\n",
       "      <td>3.935714</td>\n",
       "      <td>8.133321</td>\n",
       "      <td>6.559757</td>\n",
       "      <td>5.522244</td>\n",
       "      <td>7.599990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889395</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.020841</td>\n",
       "      <td>5.314189</td>\n",
       "      <td>3.820659</td>\n",
       "      <td>8.087557</td>\n",
       "      <td>4.653697</td>\n",
       "      <td>4.588396</td>\n",
       "      <td>4.050351</td>\n",
       "      <td>5.421180</td>\n",
       "      <td>4.346932</td>\n",
       "      <td>...</td>\n",
       "      <td>6.361815</td>\n",
       "      <td>7.686751</td>\n",
       "      <td>6.433484</td>\n",
       "      <td>4.836839</td>\n",
       "      <td>6.451500</td>\n",
       "      <td>3.903579</td>\n",
       "      <td>7.966041</td>\n",
       "      <td>8.943484</td>\n",
       "      <td>5.539704</td>\n",
       "      <td>8.019751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889396</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.014772</td>\n",
       "      <td>4.697076</td>\n",
       "      <td>4.033506</td>\n",
       "      <td>7.891401</td>\n",
       "      <td>6.844709</td>\n",
       "      <td>4.221392</td>\n",
       "      <td>4.622513</td>\n",
       "      <td>5.127564</td>\n",
       "      <td>4.334139</td>\n",
       "      <td>...</td>\n",
       "      <td>5.471339</td>\n",
       "      <td>6.793831</td>\n",
       "      <td>6.925054</td>\n",
       "      <td>5.128757</td>\n",
       "      <td>6.085150</td>\n",
       "      <td>3.662117</td>\n",
       "      <td>8.654398</td>\n",
       "      <td>6.160441</td>\n",
       "      <td>5.657073</td>\n",
       "      <td>7.444136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889397</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.740178</td>\n",
       "      <td>4.437951</td>\n",
       "      <td>3.855895</td>\n",
       "      <td>8.126038</td>\n",
       "      <td>4.865141</td>\n",
       "      <td>3.594804</td>\n",
       "      <td>4.463842</td>\n",
       "      <td>5.283395</td>\n",
       "      <td>4.636428</td>\n",
       "      <td>...</td>\n",
       "      <td>6.095071</td>\n",
       "      <td>7.400396</td>\n",
       "      <td>6.756851</td>\n",
       "      <td>5.375637</td>\n",
       "      <td>6.396927</td>\n",
       "      <td>4.001277</td>\n",
       "      <td>8.594853</td>\n",
       "      <td>6.489278</td>\n",
       "      <td>5.685232</td>\n",
       "      <td>7.856667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889398</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.241671</td>\n",
       "      <td>5.283537</td>\n",
       "      <td>3.758676</td>\n",
       "      <td>6.142525</td>\n",
       "      <td>4.503114</td>\n",
       "      <td>3.397033</td>\n",
       "      <td>4.234801</td>\n",
       "      <td>5.017120</td>\n",
       "      <td>4.186390</td>\n",
       "      <td>...</td>\n",
       "      <td>9.115309</td>\n",
       "      <td>9.574167</td>\n",
       "      <td>6.110033</td>\n",
       "      <td>5.312263</td>\n",
       "      <td>6.300677</td>\n",
       "      <td>3.478511</td>\n",
       "      <td>8.623374</td>\n",
       "      <td>6.509998</td>\n",
       "      <td>5.717293</td>\n",
       "      <td>8.357155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889399</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.906202</td>\n",
       "      <td>5.117618</td>\n",
       "      <td>3.541880</td>\n",
       "      <td>6.793005</td>\n",
       "      <td>4.300526</td>\n",
       "      <td>3.423417</td>\n",
       "      <td>5.136216</td>\n",
       "      <td>4.798390</td>\n",
       "      <td>4.334139</td>\n",
       "      <td>...</td>\n",
       "      <td>9.133508</td>\n",
       "      <td>11.364172</td>\n",
       "      <td>6.384319</td>\n",
       "      <td>5.283314</td>\n",
       "      <td>6.855975</td>\n",
       "      <td>3.398290</td>\n",
       "      <td>7.777169</td>\n",
       "      <td>5.880868</td>\n",
       "      <td>6.085382</td>\n",
       "      <td>8.687053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889400</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.924588</td>\n",
       "      <td>4.626107</td>\n",
       "      <td>3.707344</td>\n",
       "      <td>7.240634</td>\n",
       "      <td>5.701770</td>\n",
       "      <td>3.235174</td>\n",
       "      <td>3.982679</td>\n",
       "      <td>5.081208</td>\n",
       "      <td>4.635661</td>\n",
       "      <td>...</td>\n",
       "      <td>9.377168</td>\n",
       "      <td>10.879890</td>\n",
       "      <td>6.690361</td>\n",
       "      <td>4.937915</td>\n",
       "      <td>5.533938</td>\n",
       "      <td>4.324565</td>\n",
       "      <td>8.231185</td>\n",
       "      <td>6.958150</td>\n",
       "      <td>5.288436</td>\n",
       "      <td>7.894728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889401</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.115747</td>\n",
       "      <td>4.795356</td>\n",
       "      <td>4.271000</td>\n",
       "      <td>7.925575</td>\n",
       "      <td>7.060284</td>\n",
       "      <td>4.549555</td>\n",
       "      <td>4.175373</td>\n",
       "      <td>5.659138</td>\n",
       "      <td>4.893546</td>\n",
       "      <td>...</td>\n",
       "      <td>5.088151</td>\n",
       "      <td>6.443541</td>\n",
       "      <td>6.969547</td>\n",
       "      <td>5.098268</td>\n",
       "      <td>6.096490</td>\n",
       "      <td>3.755719</td>\n",
       "      <td>8.307370</td>\n",
       "      <td>6.403943</td>\n",
       "      <td>5.732134</td>\n",
       "      <td>7.646344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889402</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.281404</td>\n",
       "      <td>5.226730</td>\n",
       "      <td>3.580458</td>\n",
       "      <td>7.902166</td>\n",
       "      <td>5.513326</td>\n",
       "      <td>3.385652</td>\n",
       "      <td>3.962220</td>\n",
       "      <td>5.615050</td>\n",
       "      <td>4.255069</td>\n",
       "      <td>...</td>\n",
       "      <td>8.749590</td>\n",
       "      <td>8.727941</td>\n",
       "      <td>6.197129</td>\n",
       "      <td>4.993167</td>\n",
       "      <td>5.775538</td>\n",
       "      <td>3.620932</td>\n",
       "      <td>7.902476</td>\n",
       "      <td>7.327755</td>\n",
       "      <td>5.557686</td>\n",
       "      <td>7.360849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889403</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.524250</td>\n",
       "      <td>5.408838</td>\n",
       "      <td>3.680922</td>\n",
       "      <td>8.386771</td>\n",
       "      <td>5.026710</td>\n",
       "      <td>3.367344</td>\n",
       "      <td>3.959501</td>\n",
       "      <td>5.175938</td>\n",
       "      <td>4.306350</td>\n",
       "      <td>...</td>\n",
       "      <td>7.774954</td>\n",
       "      <td>7.072685</td>\n",
       "      <td>6.562624</td>\n",
       "      <td>4.888484</td>\n",
       "      <td>6.483230</td>\n",
       "      <td>3.454973</td>\n",
       "      <td>8.143291</td>\n",
       "      <td>7.386660</td>\n",
       "      <td>5.737469</td>\n",
       "      <td>7.829836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889404</th>\n",
       "      <td>0.0</td>\n",
       "      <td>10.645244</td>\n",
       "      <td>4.915869</td>\n",
       "      <td>9.464449</td>\n",
       "      <td>8.124039</td>\n",
       "      <td>6.675739</td>\n",
       "      <td>3.239315</td>\n",
       "      <td>4.412951</td>\n",
       "      <td>5.143401</td>\n",
       "      <td>5.136100</td>\n",
       "      <td>...</td>\n",
       "      <td>7.487605</td>\n",
       "      <td>10.408471</td>\n",
       "      <td>6.381852</td>\n",
       "      <td>5.630728</td>\n",
       "      <td>6.502749</td>\n",
       "      <td>4.485918</td>\n",
       "      <td>8.015752</td>\n",
       "      <td>7.158299</td>\n",
       "      <td>5.360036</td>\n",
       "      <td>8.098215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889405</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.662880</td>\n",
       "      <td>5.670367</td>\n",
       "      <td>3.656357</td>\n",
       "      <td>6.922179</td>\n",
       "      <td>5.129995</td>\n",
       "      <td>3.350096</td>\n",
       "      <td>3.893872</td>\n",
       "      <td>4.908408</td>\n",
       "      <td>3.751299</td>\n",
       "      <td>...</td>\n",
       "      <td>9.643950</td>\n",
       "      <td>11.090754</td>\n",
       "      <td>6.158599</td>\n",
       "      <td>5.122336</td>\n",
       "      <td>6.586682</td>\n",
       "      <td>6.512242</td>\n",
       "      <td>8.622670</td>\n",
       "      <td>7.421939</td>\n",
       "      <td>5.613927</td>\n",
       "      <td>7.616096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889406</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.858641</td>\n",
       "      <td>4.922324</td>\n",
       "      <td>3.799497</td>\n",
       "      <td>7.907427</td>\n",
       "      <td>5.948012</td>\n",
       "      <td>3.218905</td>\n",
       "      <td>4.122376</td>\n",
       "      <td>5.866440</td>\n",
       "      <td>4.372785</td>\n",
       "      <td>...</td>\n",
       "      <td>8.565045</td>\n",
       "      <td>10.572460</td>\n",
       "      <td>6.964754</td>\n",
       "      <td>4.932504</td>\n",
       "      <td>6.533675</td>\n",
       "      <td>3.772244</td>\n",
       "      <td>7.750632</td>\n",
       "      <td>6.915321</td>\n",
       "      <td>5.607471</td>\n",
       "      <td>7.239554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889407</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.249585</td>\n",
       "      <td>4.921713</td>\n",
       "      <td>3.687088</td>\n",
       "      <td>7.239692</td>\n",
       "      <td>6.021325</td>\n",
       "      <td>3.118186</td>\n",
       "      <td>4.131019</td>\n",
       "      <td>4.835022</td>\n",
       "      <td>4.383037</td>\n",
       "      <td>...</td>\n",
       "      <td>8.186844</td>\n",
       "      <td>9.327142</td>\n",
       "      <td>7.239337</td>\n",
       "      <td>5.493023</td>\n",
       "      <td>6.076990</td>\n",
       "      <td>3.839951</td>\n",
       "      <td>9.116587</td>\n",
       "      <td>6.789239</td>\n",
       "      <td>5.433131</td>\n",
       "      <td>8.094626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889408</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.062021</td>\n",
       "      <td>4.927759</td>\n",
       "      <td>3.693696</td>\n",
       "      <td>7.343096</td>\n",
       "      <td>4.576001</td>\n",
       "      <td>5.130322</td>\n",
       "      <td>3.742170</td>\n",
       "      <td>6.569115</td>\n",
       "      <td>4.163847</td>\n",
       "      <td>...</td>\n",
       "      <td>6.398014</td>\n",
       "      <td>7.037130</td>\n",
       "      <td>7.065252</td>\n",
       "      <td>5.307279</td>\n",
       "      <td>6.890025</td>\n",
       "      <td>3.742170</td>\n",
       "      <td>8.167969</td>\n",
       "      <td>7.177863</td>\n",
       "      <td>5.474702</td>\n",
       "      <td>8.221265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889409</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.209183</td>\n",
       "      <td>5.052906</td>\n",
       "      <td>3.779837</td>\n",
       "      <td>7.426246</td>\n",
       "      <td>4.158846</td>\n",
       "      <td>3.518999</td>\n",
       "      <td>4.317963</td>\n",
       "      <td>5.757417</td>\n",
       "      <td>4.412691</td>\n",
       "      <td>...</td>\n",
       "      <td>8.084736</td>\n",
       "      <td>8.967240</td>\n",
       "      <td>6.031400</td>\n",
       "      <td>5.001945</td>\n",
       "      <td>6.160809</td>\n",
       "      <td>3.774179</td>\n",
       "      <td>8.176942</td>\n",
       "      <td>7.692460</td>\n",
       "      <td>5.749370</td>\n",
       "      <td>8.064374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889410</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.862797</td>\n",
       "      <td>4.656392</td>\n",
       "      <td>4.131397</td>\n",
       "      <td>6.875909</td>\n",
       "      <td>5.869809</td>\n",
       "      <td>3.227011</td>\n",
       "      <td>3.962220</td>\n",
       "      <td>5.530469</td>\n",
       "      <td>4.334139</td>\n",
       "      <td>...</td>\n",
       "      <td>8.653143</td>\n",
       "      <td>10.241961</td>\n",
       "      <td>6.564292</td>\n",
       "      <td>5.868437</td>\n",
       "      <td>7.066195</td>\n",
       "      <td>4.165513</td>\n",
       "      <td>8.696095</td>\n",
       "      <td>6.989066</td>\n",
       "      <td>5.817303</td>\n",
       "      <td>8.593815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889411</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.150139</td>\n",
       "      <td>4.967411</td>\n",
       "      <td>3.626229</td>\n",
       "      <td>7.675416</td>\n",
       "      <td>7.084077</td>\n",
       "      <td>3.475381</td>\n",
       "      <td>3.871116</td>\n",
       "      <td>5.525234</td>\n",
       "      <td>3.874259</td>\n",
       "      <td>...</td>\n",
       "      <td>6.844702</td>\n",
       "      <td>7.705989</td>\n",
       "      <td>8.520105</td>\n",
       "      <td>6.025841</td>\n",
       "      <td>6.816121</td>\n",
       "      <td>3.739264</td>\n",
       "      <td>8.716259</td>\n",
       "      <td>5.960370</td>\n",
       "      <td>5.391619</td>\n",
       "      <td>8.128676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889412</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.242893</td>\n",
       "      <td>5.368845</td>\n",
       "      <td>3.786596</td>\n",
       "      <td>7.888413</td>\n",
       "      <td>6.641694</td>\n",
       "      <td>3.431837</td>\n",
       "      <td>4.226997</td>\n",
       "      <td>5.195542</td>\n",
       "      <td>4.288048</td>\n",
       "      <td>...</td>\n",
       "      <td>9.043329</td>\n",
       "      <td>9.629302</td>\n",
       "      <td>7.241767</td>\n",
       "      <td>5.408205</td>\n",
       "      <td>6.716485</td>\n",
       "      <td>5.881325</td>\n",
       "      <td>8.448791</td>\n",
       "      <td>6.704927</td>\n",
       "      <td>5.779138</td>\n",
       "      <td>8.718654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889413</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.758070</td>\n",
       "      <td>4.470272</td>\n",
       "      <td>3.776402</td>\n",
       "      <td>6.825230</td>\n",
       "      <td>4.258575</td>\n",
       "      <td>3.181411</td>\n",
       "      <td>3.692919</td>\n",
       "      <td>4.670817</td>\n",
       "      <td>4.271563</td>\n",
       "      <td>...</td>\n",
       "      <td>8.828972</td>\n",
       "      <td>10.906881</td>\n",
       "      <td>6.659995</td>\n",
       "      <td>4.733658</td>\n",
       "      <td>6.799251</td>\n",
       "      <td>5.979230</td>\n",
       "      <td>9.056035</td>\n",
       "      <td>6.318561</td>\n",
       "      <td>5.993546</td>\n",
       "      <td>8.386584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889414</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.278403</td>\n",
       "      <td>5.400511</td>\n",
       "      <td>3.803316</td>\n",
       "      <td>7.195041</td>\n",
       "      <td>5.360263</td>\n",
       "      <td>4.457138</td>\n",
       "      <td>3.792029</td>\n",
       "      <td>6.084044</td>\n",
       "      <td>4.184368</td>\n",
       "      <td>...</td>\n",
       "      <td>6.665486</td>\n",
       "      <td>7.576212</td>\n",
       "      <td>6.984332</td>\n",
       "      <td>5.208856</td>\n",
       "      <td>6.469962</td>\n",
       "      <td>4.271786</td>\n",
       "      <td>7.871825</td>\n",
       "      <td>6.175848</td>\n",
       "      <td>5.459600</td>\n",
       "      <td>8.079711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889415</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.272556</td>\n",
       "      <td>5.276987</td>\n",
       "      <td>4.356464</td>\n",
       "      <td>7.508384</td>\n",
       "      <td>6.652004</td>\n",
       "      <td>3.762204</td>\n",
       "      <td>4.490380</td>\n",
       "      <td>5.416895</td>\n",
       "      <td>4.937618</td>\n",
       "      <td>...</td>\n",
       "      <td>6.196848</td>\n",
       "      <td>6.455989</td>\n",
       "      <td>6.699588</td>\n",
       "      <td>4.724650</td>\n",
       "      <td>5.976636</td>\n",
       "      <td>3.688504</td>\n",
       "      <td>9.108076</td>\n",
       "      <td>6.383634</td>\n",
       "      <td>5.614055</td>\n",
       "      <td>7.378567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889416</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.600967</td>\n",
       "      <td>5.441775</td>\n",
       "      <td>5.655094</td>\n",
       "      <td>6.215682</td>\n",
       "      <td>6.177088</td>\n",
       "      <td>6.198454</td>\n",
       "      <td>7.343609</td>\n",
       "      <td>7.256729</td>\n",
       "      <td>4.878240</td>\n",
       "      <td>...</td>\n",
       "      <td>4.957284</td>\n",
       "      <td>5.229280</td>\n",
       "      <td>6.913050</td>\n",
       "      <td>4.454143</td>\n",
       "      <td>6.006164</td>\n",
       "      <td>5.309030</td>\n",
       "      <td>4.800296</td>\n",
       "      <td>5.994175</td>\n",
       "      <td>6.096445</td>\n",
       "      <td>5.492084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889417</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.657533</td>\n",
       "      <td>4.602252</td>\n",
       "      <td>3.978726</td>\n",
       "      <td>8.150064</td>\n",
       "      <td>7.062321</td>\n",
       "      <td>3.712279</td>\n",
       "      <td>4.055077</td>\n",
       "      <td>5.774108</td>\n",
       "      <td>4.427346</td>\n",
       "      <td>...</td>\n",
       "      <td>6.004489</td>\n",
       "      <td>6.528121</td>\n",
       "      <td>7.356232</td>\n",
       "      <td>4.844745</td>\n",
       "      <td>6.212191</td>\n",
       "      <td>4.069993</td>\n",
       "      <td>7.931003</td>\n",
       "      <td>7.025328</td>\n",
       "      <td>5.627089</td>\n",
       "      <td>7.758366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889418</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.614995</td>\n",
       "      <td>5.203488</td>\n",
       "      <td>4.182107</td>\n",
       "      <td>8.086973</td>\n",
       "      <td>6.452616</td>\n",
       "      <td>3.678169</td>\n",
       "      <td>3.848215</td>\n",
       "      <td>5.666855</td>\n",
       "      <td>4.338756</td>\n",
       "      <td>...</td>\n",
       "      <td>7.050558</td>\n",
       "      <td>7.606707</td>\n",
       "      <td>6.254080</td>\n",
       "      <td>4.321789</td>\n",
       "      <td>6.090306</td>\n",
       "      <td>4.565107</td>\n",
       "      <td>7.282819</td>\n",
       "      <td>6.692828</td>\n",
       "      <td>5.651303</td>\n",
       "      <td>7.725530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889419</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.211944</td>\n",
       "      <td>4.949106</td>\n",
       "      <td>3.903059</td>\n",
       "      <td>7.541484</td>\n",
       "      <td>5.194912</td>\n",
       "      <td>3.337133</td>\n",
       "      <td>4.283941</td>\n",
       "      <td>5.325193</td>\n",
       "      <td>4.007041</td>\n",
       "      <td>...</td>\n",
       "      <td>7.966364</td>\n",
       "      <td>9.782897</td>\n",
       "      <td>6.610482</td>\n",
       "      <td>5.005312</td>\n",
       "      <td>6.327394</td>\n",
       "      <td>4.002745</td>\n",
       "      <td>7.921015</td>\n",
       "      <td>6.590379</td>\n",
       "      <td>5.808518</td>\n",
       "      <td>7.964984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889420</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.638128</td>\n",
       "      <td>4.697487</td>\n",
       "      <td>3.856601</td>\n",
       "      <td>7.209844</td>\n",
       "      <td>5.040409</td>\n",
       "      <td>3.321802</td>\n",
       "      <td>4.179842</td>\n",
       "      <td>5.026732</td>\n",
       "      <td>4.083718</td>\n",
       "      <td>...</td>\n",
       "      <td>8.523008</td>\n",
       "      <td>11.243075</td>\n",
       "      <td>5.682876</td>\n",
       "      <td>4.628646</td>\n",
       "      <td>6.604868</td>\n",
       "      <td>4.078990</td>\n",
       "      <td>8.153173</td>\n",
       "      <td>6.212945</td>\n",
       "      <td>6.284363</td>\n",
       "      <td>8.239172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889421</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.302610</td>\n",
       "      <td>4.967411</td>\n",
       "      <td>3.738244</td>\n",
       "      <td>7.469303</td>\n",
       "      <td>4.631870</td>\n",
       "      <td>3.605130</td>\n",
       "      <td>4.229437</td>\n",
       "      <td>5.304737</td>\n",
       "      <td>4.240576</td>\n",
       "      <td>...</td>\n",
       "      <td>7.399932</td>\n",
       "      <td>8.374695</td>\n",
       "      <td>6.894272</td>\n",
       "      <td>5.210104</td>\n",
       "      <td>6.257470</td>\n",
       "      <td>3.747381</td>\n",
       "      <td>7.890696</td>\n",
       "      <td>6.775403</td>\n",
       "      <td>5.593404</td>\n",
       "      <td>8.386879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889422</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.382192</td>\n",
       "      <td>4.467902</td>\n",
       "      <td>3.695734</td>\n",
       "      <td>7.554861</td>\n",
       "      <td>4.431435</td>\n",
       "      <td>3.273938</td>\n",
       "      <td>4.215699</td>\n",
       "      <td>6.278949</td>\n",
       "      <td>4.060323</td>\n",
       "      <td>...</td>\n",
       "      <td>8.002621</td>\n",
       "      <td>9.911132</td>\n",
       "      <td>6.410429</td>\n",
       "      <td>5.006752</td>\n",
       "      <td>5.924654</td>\n",
       "      <td>3.798240</td>\n",
       "      <td>7.555687</td>\n",
       "      <td>7.361366</td>\n",
       "      <td>5.630627</td>\n",
       "      <td>7.604333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2889423</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.806563</td>\n",
       "      <td>4.628504</td>\n",
       "      <td>3.464957</td>\n",
       "      <td>6.911613</td>\n",
       "      <td>6.644617</td>\n",
       "      <td>6.351684</td>\n",
       "      <td>3.751880</td>\n",
       "      <td>6.069876</td>\n",
       "      <td>4.541674</td>\n",
       "      <td>...</td>\n",
       "      <td>6.845626</td>\n",
       "      <td>8.437546</td>\n",
       "      <td>7.069987</td>\n",
       "      <td>5.398564</td>\n",
       "      <td>6.364042</td>\n",
       "      <td>4.433770</td>\n",
       "      <td>8.743453</td>\n",
       "      <td>5.754192</td>\n",
       "      <td>5.498746</td>\n",
       "      <td>8.283418</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43 rows Ã— 17858 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Adrenocortical Cancer       A1BG  A1BG-AS1      A1CF       A2M  \\\n",
       "GSM2889381                    0.0   5.523549  5.312258  3.603395  8.439460   \n",
       "GSM2889382                    0.0   6.286303  5.636363  3.772548  6.234144   \n",
       "GSM2889383                    0.0   4.722706  4.377456  3.782952  8.099240   \n",
       "GSM2889384                    0.0   6.001011  4.989505  3.666064  7.572022   \n",
       "GSM2889385                    0.0   4.748103  4.427621  4.272122  8.578123   \n",
       "GSM2889386                    0.0   5.784339  4.967411  4.168318  7.410907   \n",
       "GSM2889387                    0.0   5.073090  6.596631  3.925859  7.690870   \n",
       "GSM2889388                    0.0   5.322978  4.705341  4.190043  7.885234   \n",
       "GSM2889389                    0.0   5.552850  4.475880  4.335184  7.469755   \n",
       "GSM2889390                    0.0  10.028951  6.932284  9.638197  9.430534   \n",
       "GSM2889391                    0.0   5.417395  5.015280  3.714774  8.120436   \n",
       "GSM2889392                    0.0   4.975241  4.258458  3.563695  7.553463   \n",
       "GSM2889393                    0.0   6.219072  4.967411  3.835905  8.228831   \n",
       "GSM2889394                    0.0   5.911014  4.760648  3.614136  8.089677   \n",
       "GSM2889395                    0.0   6.020841  5.314189  3.820659  8.087557   \n",
       "GSM2889396                    0.0   5.014772  4.697076  4.033506  7.891401   \n",
       "GSM2889397                    0.0   5.740178  4.437951  3.855895  8.126038   \n",
       "GSM2889398                    0.0   6.241671  5.283537  3.758676  6.142525   \n",
       "GSM2889399                    0.0   6.906202  5.117618  3.541880  6.793005   \n",
       "GSM2889400                    0.0   5.924588  4.626107  3.707344  7.240634   \n",
       "GSM2889401                    0.0   4.115747  4.795356  4.271000  7.925575   \n",
       "GSM2889402                    0.0   7.281404  5.226730  3.580458  7.902166   \n",
       "GSM2889403                    0.0   5.524250  5.408838  3.680922  8.386771   \n",
       "GSM2889404                    0.0  10.645244  4.915869  9.464449  8.124039   \n",
       "GSM2889405                    0.0   6.662880  5.670367  3.656357  6.922179   \n",
       "GSM2889406                    0.0   5.858641  4.922324  3.799497  7.907427   \n",
       "GSM2889407                    0.0   6.249585  4.921713  3.687088  7.239692   \n",
       "GSM2889408                    0.0   6.062021  4.927759  3.693696  7.343096   \n",
       "GSM2889409                    0.0   6.209183  5.052906  3.779837  7.426246   \n",
       "GSM2889410                    0.0   4.862797  4.656392  4.131397  6.875909   \n",
       "GSM2889411                    0.0   6.150139  4.967411  3.626229  7.675416   \n",
       "GSM2889412                    0.0   8.242893  5.368845  3.786596  7.888413   \n",
       "GSM2889413                    0.0   5.758070  4.470272  3.776402  6.825230   \n",
       "GSM2889414                    0.0   6.278403  5.400511  3.803316  7.195041   \n",
       "GSM2889415                    0.0   6.272556  5.276987  4.356464  7.508384   \n",
       "GSM2889416                    0.0   4.600967  5.441775  5.655094  6.215682   \n",
       "GSM2889417                    0.0   4.657533  4.602252  3.978726  8.150064   \n",
       "GSM2889418                    0.0   5.614995  5.203488  4.182107  8.086973   \n",
       "GSM2889419                    0.0   6.211944  4.949106  3.903059  7.541484   \n",
       "GSM2889420                    1.0   5.638128  4.697487  3.856601  7.209844   \n",
       "GSM2889421                    0.0   6.302610  4.967411  3.738244  7.469303   \n",
       "GSM2889422                    0.0   5.382192  4.467902  3.695734  7.554861   \n",
       "GSM2889423                    0.0   5.806563  4.628504  3.464957  6.911613   \n",
       "\n",
       "             A2M-AS1     A2ML1     A2MP1    A4GALT     A4GNT  ...    ZWILCH  \\\n",
       "GSM2889381  6.784101  3.275863  3.821555  5.833153  3.929618  ...  7.404137   \n",
       "GSM2889382  4.090659  3.355778  3.929229  5.390436  4.152816  ...  9.315457   \n",
       "GSM2889383  5.154216  3.262292  4.774103  4.825479  4.381623  ...  7.241154   \n",
       "GSM2889384  7.076977  3.777149  4.317872  5.557962  4.262175  ...  6.538005   \n",
       "GSM2889385  6.310073  3.301956  4.272148  5.374137  4.317426  ...  7.874055   \n",
       "GSM2889386  5.152857  3.246321  3.977649  5.160725  6.856612  ...  7.595532   \n",
       "GSM2889387  5.589711  3.595074  3.510832  7.146452  3.627231  ...  5.044171   \n",
       "GSM2889388  4.979747  3.422809  3.838083  5.613246  4.328624  ...  6.942579   \n",
       "GSM2889389  4.542321  3.403456  4.072234  6.805424  5.046665  ...  6.345089   \n",
       "GSM2889390  5.548948  3.741756  4.228405  8.081186  4.556953  ...  3.688915   \n",
       "GSM2889391  4.947901  3.418168  4.211228  5.074946  4.096684  ...  8.363608   \n",
       "GSM2889392  4.995097  3.347005  3.990605  4.632107  4.108716  ...  7.753695   \n",
       "GSM2889393  5.810245  3.611951  4.472261  5.971737  4.770631  ...  6.045949   \n",
       "GSM2889394  4.868123  3.388084  3.817156  5.588154  4.077890  ...  6.337353   \n",
       "GSM2889395  4.653697  4.588396  4.050351  5.421180  4.346932  ...  6.361815   \n",
       "GSM2889396  6.844709  4.221392  4.622513  5.127564  4.334139  ...  5.471339   \n",
       "GSM2889397  4.865141  3.594804  4.463842  5.283395  4.636428  ...  6.095071   \n",
       "GSM2889398  4.503114  3.397033  4.234801  5.017120  4.186390  ...  9.115309   \n",
       "GSM2889399  4.300526  3.423417  5.136216  4.798390  4.334139  ...  9.133508   \n",
       "GSM2889400  5.701770  3.235174  3.982679  5.081208  4.635661  ...  9.377168   \n",
       "GSM2889401  7.060284  4.549555  4.175373  5.659138  4.893546  ...  5.088151   \n",
       "GSM2889402  5.513326  3.385652  3.962220  5.615050  4.255069  ...  8.749590   \n",
       "GSM2889403  5.026710  3.367344  3.959501  5.175938  4.306350  ...  7.774954   \n",
       "GSM2889404  6.675739  3.239315  4.412951  5.143401  5.136100  ...  7.487605   \n",
       "GSM2889405  5.129995  3.350096  3.893872  4.908408  3.751299  ...  9.643950   \n",
       "GSM2889406  5.948012  3.218905  4.122376  5.866440  4.372785  ...  8.565045   \n",
       "GSM2889407  6.021325  3.118186  4.131019  4.835022  4.383037  ...  8.186844   \n",
       "GSM2889408  4.576001  5.130322  3.742170  6.569115  4.163847  ...  6.398014   \n",
       "GSM2889409  4.158846  3.518999  4.317963  5.757417  4.412691  ...  8.084736   \n",
       "GSM2889410  5.869809  3.227011  3.962220  5.530469  4.334139  ...  8.653143   \n",
       "GSM2889411  7.084077  3.475381  3.871116  5.525234  3.874259  ...  6.844702   \n",
       "GSM2889412  6.641694  3.431837  4.226997  5.195542  4.288048  ...  9.043329   \n",
       "GSM2889413  4.258575  3.181411  3.692919  4.670817  4.271563  ...  8.828972   \n",
       "GSM2889414  5.360263  4.457138  3.792029  6.084044  4.184368  ...  6.665486   \n",
       "GSM2889415  6.652004  3.762204  4.490380  5.416895  4.937618  ...  6.196848   \n",
       "GSM2889416  6.177088  6.198454  7.343609  7.256729  4.878240  ...  4.957284   \n",
       "GSM2889417  7.062321  3.712279  4.055077  5.774108  4.427346  ...  6.004489   \n",
       "GSM2889418  6.452616  3.678169  3.848215  5.666855  4.338756  ...  7.050558   \n",
       "GSM2889419  5.194912  3.337133  4.283941  5.325193  4.007041  ...  7.966364   \n",
       "GSM2889420  5.040409  3.321802  4.179842  5.026732  4.083718  ...  8.523008   \n",
       "GSM2889421  4.631870  3.605130  4.229437  5.304737  4.240576  ...  7.399932   \n",
       "GSM2889422  4.431435  3.273938  4.215699  6.278949  4.060323  ...  8.002621   \n",
       "GSM2889423  6.644617  6.351684  3.751880  6.069876  4.541674  ...  6.845626   \n",
       "\n",
       "                ZWINT      ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B  \\\n",
       "GSM2889381   8.066233  7.023525  5.423627  6.130173  3.875443  8.324267   \n",
       "GSM2889382  11.280892  6.738129  5.336899  6.629826  6.901420  8.563336   \n",
       "GSM2889383   7.623172  7.457557  5.477725  6.049950  3.827137  8.407677   \n",
       "GSM2889384   6.921254  7.068052  5.379503  6.370581  3.950170  8.419080   \n",
       "GSM2889385   7.974063  6.767503  4.873842  5.980389  3.911654  8.368285   \n",
       "GSM2889386   9.056683  6.624460  5.334879  5.982906  3.895866  7.756354   \n",
       "GSM2889387   2.995060  5.120191  6.015022  6.120873  2.871190  7.599136   \n",
       "GSM2889388   7.837928  6.339747  4.788353  6.381847  4.332635  7.606243   \n",
       "GSM2889389   8.777369  7.087475  5.236264  5.944496  3.780255  7.776769   \n",
       "GSM2889390   2.822378  3.483630  5.673570  5.903220  4.327800  4.288961   \n",
       "GSM2889391   9.731977  6.729941  5.507374  6.100880  4.129861  9.114931   \n",
       "GSM2889392  10.239855  7.474382  5.820885  6.783218  3.810572  8.988548   \n",
       "GSM2889393   7.027758  5.806202  4.721118  6.053315  4.553429  8.015609   \n",
       "GSM2889394   8.070385  7.853956  5.449738  6.900473  3.935714  8.133321   \n",
       "GSM2889395   7.686751  6.433484  4.836839  6.451500  3.903579  7.966041   \n",
       "GSM2889396   6.793831  6.925054  5.128757  6.085150  3.662117  8.654398   \n",
       "GSM2889397   7.400396  6.756851  5.375637  6.396927  4.001277  8.594853   \n",
       "GSM2889398   9.574167  6.110033  5.312263  6.300677  3.478511  8.623374   \n",
       "GSM2889399  11.364172  6.384319  5.283314  6.855975  3.398290  7.777169   \n",
       "GSM2889400  10.879890  6.690361  4.937915  5.533938  4.324565  8.231185   \n",
       "GSM2889401   6.443541  6.969547  5.098268  6.096490  3.755719  8.307370   \n",
       "GSM2889402   8.727941  6.197129  4.993167  5.775538  3.620932  7.902476   \n",
       "GSM2889403   7.072685  6.562624  4.888484  6.483230  3.454973  8.143291   \n",
       "GSM2889404  10.408471  6.381852  5.630728  6.502749  4.485918  8.015752   \n",
       "GSM2889405  11.090754  6.158599  5.122336  6.586682  6.512242  8.622670   \n",
       "GSM2889406  10.572460  6.964754  4.932504  6.533675  3.772244  7.750632   \n",
       "GSM2889407   9.327142  7.239337  5.493023  6.076990  3.839951  9.116587   \n",
       "GSM2889408   7.037130  7.065252  5.307279  6.890025  3.742170  8.167969   \n",
       "GSM2889409   8.967240  6.031400  5.001945  6.160809  3.774179  8.176942   \n",
       "GSM2889410  10.241961  6.564292  5.868437  7.066195  4.165513  8.696095   \n",
       "GSM2889411   7.705989  8.520105  6.025841  6.816121  3.739264  8.716259   \n",
       "GSM2889412   9.629302  7.241767  5.408205  6.716485  5.881325  8.448791   \n",
       "GSM2889413  10.906881  6.659995  4.733658  6.799251  5.979230  9.056035   \n",
       "GSM2889414   7.576212  6.984332  5.208856  6.469962  4.271786  7.871825   \n",
       "GSM2889415   6.455989  6.699588  4.724650  5.976636  3.688504  9.108076   \n",
       "GSM2889416   5.229280  6.913050  4.454143  6.006164  5.309030  4.800296   \n",
       "GSM2889417   6.528121  7.356232  4.844745  6.212191  4.069993  7.931003   \n",
       "GSM2889418   7.606707  6.254080  4.321789  6.090306  4.565107  7.282819   \n",
       "GSM2889419   9.782897  6.610482  5.005312  6.327394  4.002745  7.921015   \n",
       "GSM2889420  11.243075  5.682876  4.628646  6.604868  4.078990  8.153173   \n",
       "GSM2889421   8.374695  6.894272  5.210104  6.257470  3.747381  7.890696   \n",
       "GSM2889422   9.911132  6.410429  5.006752  5.924654  3.798240  7.555687   \n",
       "GSM2889423   8.437546  7.069987  5.398564  6.364042  4.433770  8.743453   \n",
       "\n",
       "                 ZYX     ZZEF1      ZZZ3  \n",
       "GSM2889381  6.495728  5.451428  8.522046  \n",
       "GSM2889382  7.109820  6.005562  8.770399  \n",
       "GSM2889383  6.811351  5.749764  8.118409  \n",
       "GSM2889384  7.635857  5.674700  8.082138  \n",
       "GSM2889385  8.528983  5.761255  7.635416  \n",
       "GSM2889386  7.296132  5.862683  8.030460  \n",
       "GSM2889387  9.704778  5.487030  5.420636  \n",
       "GSM2889388  7.226106  5.811785  7.711138  \n",
       "GSM2889389  6.748325  5.731223  6.964051  \n",
       "GSM2889390  9.603028  6.066631  4.419665  \n",
       "GSM2889391  7.631545  5.577168  8.143149  \n",
       "GSM2889392  6.962456  5.607036  7.965766  \n",
       "GSM2889393  7.105865  5.686411  7.388064  \n",
       "GSM2889394  6.559757  5.522244  7.599990  \n",
       "GSM2889395  8.943484  5.539704  8.019751  \n",
       "GSM2889396  6.160441  5.657073  7.444136  \n",
       "GSM2889397  6.489278  5.685232  7.856667  \n",
       "GSM2889398  6.509998  5.717293  8.357155  \n",
       "GSM2889399  5.880868  6.085382  8.687053  \n",
       "GSM2889400  6.958150  5.288436  7.894728  \n",
       "GSM2889401  6.403943  5.732134  7.646344  \n",
       "GSM2889402  7.327755  5.557686  7.360849  \n",
       "GSM2889403  7.386660  5.737469  7.829836  \n",
       "GSM2889404  7.158299  5.360036  8.098215  \n",
       "GSM2889405  7.421939  5.613927  7.616096  \n",
       "GSM2889406  6.915321  5.607471  7.239554  \n",
       "GSM2889407  6.789239  5.433131  8.094626  \n",
       "GSM2889408  7.177863  5.474702  8.221265  \n",
       "GSM2889409  7.692460  5.749370  8.064374  \n",
       "GSM2889410  6.989066  5.817303  8.593815  \n",
       "GSM2889411  5.960370  5.391619  8.128676  \n",
       "GSM2889412  6.704927  5.779138  8.718654  \n",
       "GSM2889413  6.318561  5.993546  8.386584  \n",
       "GSM2889414  6.175848  5.459600  8.079711  \n",
       "GSM2889415  6.383634  5.614055  7.378567  \n",
       "GSM2889416  5.994175  6.096445  5.492084  \n",
       "GSM2889417  7.025328  5.627089  7.758366  \n",
       "GSM2889418  6.692828  5.651303  7.725530  \n",
       "GSM2889419  6.590379  5.808518  7.964984  \n",
       "GSM2889420  6.212945  6.284363  8.239172  \n",
       "GSM2889421  6.775403  5.593404  8.386879  \n",
       "GSM2889422  7.361366  5.630627  7.604333  \n",
       "GSM2889423  5.754192  5.498746  8.283418  \n",
       "\n",
       "[43 rows x 17858 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 43 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 1 occurrences. This represents 2.33% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Expression profiling by array of 44 adrenocortical carcinomas\"\n",
      "!Series_summary\t\"Gene expression profiles of adrenocortical carcinomas were analyzed using Affymetrix Human Gene 2.0 ST Array to identify homogeneous molecular subgroups\"\n",
      "!Series_overall_design\t\"Gene expression profiles of 44 adrenocortical carcinomas were hybridized using Affymetrix Human Gene 2.0 ST Array\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1196511</th>\n",
       "      <th>GSM1196512</th>\n",
       "      <th>GSM1196513</th>\n",
       "      <th>GSM1196514</th>\n",
       "      <th>GSM1196515</th>\n",
       "      <th>GSM1196516</th>\n",
       "      <th>GSM1196517</th>\n",
       "      <th>GSM1196518</th>\n",
       "      <th>GSM1196519</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196545</th>\n",
       "      <th>GSM1196546</th>\n",
       "      <th>GSM1196547</th>\n",
       "      <th>GSM1196548</th>\n",
       "      <th>GSM1196549</th>\n",
       "      <th>GSM1196550</th>\n",
       "      <th>GSM1196551</th>\n",
       "      <th>GSM1196552</th>\n",
       "      <th>GSM1196553</th>\n",
       "      <th>GSM1196554</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age (years): 70</td>\n",
       "      <td>age (years): 26</td>\n",
       "      <td>age (years): 53</td>\n",
       "      <td>age (years): 73</td>\n",
       "      <td>age (years): 15</td>\n",
       "      <td>age (years): 51</td>\n",
       "      <td>age (years): 63</td>\n",
       "      <td>age (years): 26</td>\n",
       "      <td>age (years): 29</td>\n",
       "      <td>...</td>\n",
       "      <td>age (years): 79</td>\n",
       "      <td>age (years): 28</td>\n",
       "      <td>age (years): 40</td>\n",
       "      <td>age (years): 44</td>\n",
       "      <td>age (years): 28</td>\n",
       "      <td>age (years): 53</td>\n",
       "      <td>age (years): 28</td>\n",
       "      <td>age (years): 52</td>\n",
       "      <td>age (years): 30</td>\n",
       "      <td>age (years): 46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "      <td>cell type: Adrenocortical carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM1196511  \\\n",
       "0  !Sample_characteristics_ch1                      age (years): 70   \n",
       "1  !Sample_characteristics_ch1                            gender: F   \n",
       "2  !Sample_characteristics_ch1  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196512                           GSM1196513  \\\n",
       "0                      age (years): 26                      age (years): 53   \n",
       "1                            gender: F                            gender: F   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196514                           GSM1196515  \\\n",
       "0                      age (years): 73                      age (years): 15   \n",
       "1                            gender: M                            gender: F   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196516                           GSM1196517  \\\n",
       "0                      age (years): 51                      age (years): 63   \n",
       "1                            gender: F                            gender: M   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196518                           GSM1196519  \\\n",
       "0                      age (years): 26                      age (years): 29   \n",
       "1                            gender: F                            gender: M   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "   ...                           GSM1196545  \\\n",
       "0  ...                      age (years): 79   \n",
       "1  ...                            gender: F   \n",
       "2  ...  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196546                           GSM1196547  \\\n",
       "0                      age (years): 28                      age (years): 40   \n",
       "1                            gender: F                            gender: F   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196548                           GSM1196549  \\\n",
       "0                      age (years): 44                      age (years): 28   \n",
       "1                            gender: F                            gender: F   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196550                           GSM1196551  \\\n",
       "0                      age (years): 53                      age (years): 28   \n",
       "1                            gender: F                            gender: M   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196552                           GSM1196553  \\\n",
       "0                      age (years): 52                      age (years): 30   \n",
       "1                            gender: M                            gender: F   \n",
       "2  cell type: Adrenocortical carcinoma  cell type: Adrenocortical carcinoma   \n",
       "\n",
       "                            GSM1196554  \n",
       "0                      age (years): 46  \n",
       "1                            gender: F  \n",
       "2  cell type: Adrenocortical carcinoma  \n",
       "\n",
       "[3 rows x 45 columns]"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No gene mapping\n",
    "cohort = accession_num = \"GSE49278\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'cell type: Adrenocortical carcinoma'], dtype=object)"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = 0\n",
    "gender_row = 1\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'cell type: Adrenocortical carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196511</th>\n",
       "      <th>GSM1196512</th>\n",
       "      <th>GSM1196513</th>\n",
       "      <th>GSM1196514</th>\n",
       "      <th>GSM1196515</th>\n",
       "      <th>GSM1196516</th>\n",
       "      <th>GSM1196517</th>\n",
       "      <th>GSM1196518</th>\n",
       "      <th>GSM1196519</th>\n",
       "      <th>GSM1196520</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196545</th>\n",
       "      <th>GSM1196546</th>\n",
       "      <th>GSM1196547</th>\n",
       "      <th>GSM1196548</th>\n",
       "      <th>GSM1196549</th>\n",
       "      <th>GSM1196550</th>\n",
       "      <th>GSM1196551</th>\n",
       "      <th>GSM1196552</th>\n",
       "      <th>GSM1196553</th>\n",
       "      <th>GSM1196554</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>70</td>\n",
       "      <td>26</td>\n",
       "      <td>53</td>\n",
       "      <td>73</td>\n",
       "      <td>15</td>\n",
       "      <td>51</td>\n",
       "      <td>63</td>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>79</td>\n",
       "      <td>...</td>\n",
       "      <td>79</td>\n",
       "      <td>28</td>\n",
       "      <td>40</td>\n",
       "      <td>44</td>\n",
       "      <td>28</td>\n",
       "      <td>53</td>\n",
       "      <td>28</td>\n",
       "      <td>52</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1196511  GSM1196512  GSM1196513  GSM1196514  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Age                            70          26          53          73   \n",
       "Gender                          1           1           1           0   \n",
       "\n",
       "                       GSM1196515  GSM1196516  GSM1196517  GSM1196518  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Age                            15          51          63          26   \n",
       "Gender                          1           1           0           1   \n",
       "\n",
       "                       GSM1196519  GSM1196520  ...  GSM1196545  GSM1196546  \\\n",
       "Adrenocortical Cancer           1           1  ...           1           1   \n",
       "Age                            29          79  ...          79          28   \n",
       "Gender                          0           1  ...           1           1   \n",
       "\n",
       "                       GSM1196547  GSM1196548  GSM1196549  GSM1196550  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "Age                            40          44          28          53   \n",
       "Gender                          1           1           1           1   \n",
       "\n",
       "                       GSM1196551  GSM1196552  GSM1196553  GSM1196554  \n",
       "Adrenocortical Cancer           1           1           1           1  \n",
       "Age                            28          52          30          46  \n",
       "Gender                          0           0           1           1  \n",
       "\n",
       "[3 rows x 44 columns]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1196511</th>\n",
       "      <th>GSM1196512</th>\n",
       "      <th>GSM1196513</th>\n",
       "      <th>GSM1196514</th>\n",
       "      <th>GSM1196515</th>\n",
       "      <th>GSM1196516</th>\n",
       "      <th>GSM1196517</th>\n",
       "      <th>GSM1196518</th>\n",
       "      <th>GSM1196519</th>\n",
       "      <th>GSM1196520</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1196545</th>\n",
       "      <th>GSM1196546</th>\n",
       "      <th>GSM1196547</th>\n",
       "      <th>GSM1196548</th>\n",
       "      <th>GSM1196549</th>\n",
       "      <th>GSM1196550</th>\n",
       "      <th>GSM1196551</th>\n",
       "      <th>GSM1196552</th>\n",
       "      <th>GSM1196553</th>\n",
       "      <th>GSM1196554</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16650001</th>\n",
       "      <td>3.114460</td>\n",
       "      <td>2.761934</td>\n",
       "      <td>3.191700</td>\n",
       "      <td>2.981038</td>\n",
       "      <td>3.113831</td>\n",
       "      <td>2.687413</td>\n",
       "      <td>3.468881</td>\n",
       "      <td>2.411585</td>\n",
       "      <td>3.761057</td>\n",
       "      <td>2.974074</td>\n",
       "      <td>...</td>\n",
       "      <td>2.440173</td>\n",
       "      <td>2.954796</td>\n",
       "      <td>3.445082</td>\n",
       "      <td>3.388275</td>\n",
       "      <td>2.450530</td>\n",
       "      <td>2.293126</td>\n",
       "      <td>3.136449</td>\n",
       "      <td>2.748609</td>\n",
       "      <td>3.587116</td>\n",
       "      <td>3.194252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16650003</th>\n",
       "      <td>2.070307</td>\n",
       "      <td>1.831540</td>\n",
       "      <td>2.303189</td>\n",
       "      <td>2.430376</td>\n",
       "      <td>1.507325</td>\n",
       "      <td>2.382929</td>\n",
       "      <td>2.808405</td>\n",
       "      <td>2.031501</td>\n",
       "      <td>2.797925</td>\n",
       "      <td>2.567698</td>\n",
       "      <td>...</td>\n",
       "      <td>1.833832</td>\n",
       "      <td>2.025689</td>\n",
       "      <td>2.493108</td>\n",
       "      <td>2.313400</td>\n",
       "      <td>1.594192</td>\n",
       "      <td>2.106784</td>\n",
       "      <td>3.733405</td>\n",
       "      <td>2.427485</td>\n",
       "      <td>3.297156</td>\n",
       "      <td>1.920650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16650005</th>\n",
       "      <td>2.532754</td>\n",
       "      <td>3.371765</td>\n",
       "      <td>2.264750</td>\n",
       "      <td>2.647668</td>\n",
       "      <td>2.559651</td>\n",
       "      <td>3.508271</td>\n",
       "      <td>1.959297</td>\n",
       "      <td>2.764491</td>\n",
       "      <td>2.655120</td>\n",
       "      <td>1.712738</td>\n",
       "      <td>...</td>\n",
       "      <td>2.165530</td>\n",
       "      <td>4.164357</td>\n",
       "      <td>3.455904</td>\n",
       "      <td>4.223868</td>\n",
       "      <td>2.515237</td>\n",
       "      <td>2.956488</td>\n",
       "      <td>3.047515</td>\n",
       "      <td>1.870629</td>\n",
       "      <td>2.264684</td>\n",
       "      <td>4.401433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16650007</th>\n",
       "      <td>1.968311</td>\n",
       "      <td>2.229541</td>\n",
       "      <td>1.762466</td>\n",
       "      <td>2.827752</td>\n",
       "      <td>1.626150</td>\n",
       "      <td>2.184046</td>\n",
       "      <td>1.214179</td>\n",
       "      <td>1.664709</td>\n",
       "      <td>1.559880</td>\n",
       "      <td>2.373817</td>\n",
       "      <td>...</td>\n",
       "      <td>3.381329</td>\n",
       "      <td>2.235444</td>\n",
       "      <td>2.027248</td>\n",
       "      <td>1.226888</td>\n",
       "      <td>1.948129</td>\n",
       "      <td>1.840212</td>\n",
       "      <td>2.096553</td>\n",
       "      <td>2.489499</td>\n",
       "      <td>2.316459</td>\n",
       "      <td>1.641595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16650009</th>\n",
       "      <td>1.418189</td>\n",
       "      <td>1.314710</td>\n",
       "      <td>1.571579</td>\n",
       "      <td>1.233351</td>\n",
       "      <td>1.753973</td>\n",
       "      <td>1.033928</td>\n",
       "      <td>1.259945</td>\n",
       "      <td>1.239220</td>\n",
       "      <td>1.104874</td>\n",
       "      <td>1.285327</td>\n",
       "      <td>...</td>\n",
       "      <td>1.765765</td>\n",
       "      <td>0.843827</td>\n",
       "      <td>1.600953</td>\n",
       "      <td>1.369317</td>\n",
       "      <td>0.956487</td>\n",
       "      <td>1.137052</td>\n",
       "      <td>1.658009</td>\n",
       "      <td>1.689291</td>\n",
       "      <td>1.196682</td>\n",
       "      <td>1.994568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM1196511  GSM1196512  GSM1196513  GSM1196514  GSM1196515  \\\n",
       "ID                                                                     \n",
       "16650001    3.114460    2.761934    3.191700    2.981038    3.113831   \n",
       "16650003    2.070307    1.831540    2.303189    2.430376    1.507325   \n",
       "16650005    2.532754    3.371765    2.264750    2.647668    2.559651   \n",
       "16650007    1.968311    2.229541    1.762466    2.827752    1.626150   \n",
       "16650009    1.418189    1.314710    1.571579    1.233351    1.753973   \n",
       "\n",
       "          GSM1196516  GSM1196517  GSM1196518  GSM1196519  GSM1196520  ...  \\\n",
       "ID                                                                    ...   \n",
       "16650001    2.687413    3.468881    2.411585    3.761057    2.974074  ...   \n",
       "16650003    2.382929    2.808405    2.031501    2.797925    2.567698  ...   \n",
       "16650005    3.508271    1.959297    2.764491    2.655120    1.712738  ...   \n",
       "16650007    2.184046    1.214179    1.664709    1.559880    2.373817  ...   \n",
       "16650009    1.033928    1.259945    1.239220    1.104874    1.285327  ...   \n",
       "\n",
       "          GSM1196545  GSM1196546  GSM1196547  GSM1196548  GSM1196549  \\\n",
       "ID                                                                     \n",
       "16650001    2.440173    2.954796    3.445082    3.388275    2.450530   \n",
       "16650003    1.833832    2.025689    2.493108    2.313400    1.594192   \n",
       "16650005    2.165530    4.164357    3.455904    4.223868    2.515237   \n",
       "16650007    3.381329    2.235444    2.027248    1.226888    1.948129   \n",
       "16650009    1.765765    0.843827    1.600953    1.369317    0.956487   \n",
       "\n",
       "          GSM1196550  GSM1196551  GSM1196552  GSM1196553  GSM1196554  \n",
       "ID                                                                    \n",
       "16650001    2.293126    3.136449    2.748609    3.587116    3.194252  \n",
       "16650003    2.106784    3.733405    2.427485    3.297156    1.920650  \n",
       "16650005    2.956488    3.047515    1.870629    2.264684    4.401433  \n",
       "16650007    1.840212    2.096553    2.489499    2.316459    1.641595  \n",
       "16650009    1.137052    1.658009    1.689291    1.196682    1.994568  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['16657436', '16657440', '16657445', '16657447', '16657450'], 'RANGE_STRAND': ['+', '+', '+', '+', '+'], 'RANGE_START': [12190.0, 29554.0, 69091.0, 160446.0, 317811.0], 'RANGE_END': [13639.0, 31109.0, 70008.0, 161525.0, 328581.0], 'total_probes': [25.0, 28.0, 8.0, 13.0, 36.0], 'GB_ACC': ['NR_046018', nan, nan, nan, 'NR_024368'], 'SPOT_ID': ['chr1:12190-13639', 'chr1:29554-31109', 'chr1:69091-70008', 'chr1:160446-161525', 'chr1:317811-328581'], 'RANGE_GB': ['NC_000001.10', 'NC_000001.10', 'NC_000001.10', 'NC_000001.10', 'NC_000001.10']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['16650001', '16650003', '16650005', '16650007', '16650009', '16650011', '16650013', '16650015', '16650017', '16650019', '16650021', '16650023', '16650025', '16650027', '16650029', '16650031', '16650033', '16650035', '16650037', '16650041']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['16657436', '16657440', '16657445', '16657447', '16657450'], 'RANGE_STRAND': ['+', '+', '+', '+', '+'], 'RANGE_START': [12190.0, 29554.0, 69091.0, 160446.0, 317811.0], 'RANGE_END': [13639.0, 31109.0, 70008.0, 161525.0, 328581.0], 'total_probes': [25.0, 28.0, 8.0, 13.0, 36.0], 'GB_ACC': ['NR_046018', nan, nan, nan, 'NR_024368'], 'SPOT_ID': ['chr1:12190-13639', 'chr1:29554-31109', 'chr1:69091-70008', 'chr1:160446-161525', 'chr1:317811-328581'], 'RANGE_GB': ['NC_000001.10', 'NC_000001.10', 'NC_000001.10', 'NC_000001.10', 'NC_000001.10']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>RANGE_STRAND</th>\n",
       "      <th>RANGE_START</th>\n",
       "      <th>RANGE_END</th>\n",
       "      <th>total_probes</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>RANGE_GB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16657436</td>\n",
       "      <td>+</td>\n",
       "      <td>12190.0</td>\n",
       "      <td>13639.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>chr1:12190-13639</td>\n",
       "      <td>NC_000001.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16657440</td>\n",
       "      <td>+</td>\n",
       "      <td>29554.0</td>\n",
       "      <td>31109.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:29554-31109</td>\n",
       "      <td>NC_000001.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16657445</td>\n",
       "      <td>+</td>\n",
       "      <td>69091.0</td>\n",
       "      <td>70008.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:69091-70008</td>\n",
       "      <td>NC_000001.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16657447</td>\n",
       "      <td>+</td>\n",
       "      <td>160446.0</td>\n",
       "      <td>161525.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr1:160446-161525</td>\n",
       "      <td>NC_000001.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16657450</td>\n",
       "      <td>+</td>\n",
       "      <td>317811.0</td>\n",
       "      <td>328581.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>NR_024368</td>\n",
       "      <td>chr1:317811-328581</td>\n",
       "      <td>NC_000001.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413168</th>\n",
       "      <td>17127713</td>\n",
       "      <td>5.932226625</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413169</th>\n",
       "      <td>17127715</td>\n",
       "      <td>13.32740757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413170</th>\n",
       "      <td>17127717</td>\n",
       "      <td>4.268049934</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413171</th>\n",
       "      <td>17127719</td>\n",
       "      <td>13.16387444</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413172</th>\n",
       "      <td>17127721</td>\n",
       "      <td>5.588639967</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2413173 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID RANGE_STRAND  RANGE_START  RANGE_END  total_probes  \\\n",
       "0        16657436            +      12190.0    13639.0          25.0   \n",
       "1        16657440            +      29554.0    31109.0          28.0   \n",
       "2        16657445            +      69091.0    70008.0           8.0   \n",
       "3        16657447            +     160446.0   161525.0          13.0   \n",
       "4        16657450            +     317811.0   328581.0          36.0   \n",
       "...           ...          ...          ...        ...           ...   \n",
       "2413168  17127713  5.932226625          NaN        NaN           NaN   \n",
       "2413169  17127715  13.32740757          NaN        NaN           NaN   \n",
       "2413170  17127717  4.268049934          NaN        NaN           NaN   \n",
       "2413171  17127719  13.16387444          NaN        NaN           NaN   \n",
       "2413172  17127721  5.588639967          NaN        NaN           NaN   \n",
       "\n",
       "            GB_ACC             SPOT_ID      RANGE_GB  \n",
       "0        NR_046018    chr1:12190-13639  NC_000001.10  \n",
       "1              NaN    chr1:29554-31109  NC_000001.10  \n",
       "2              NaN    chr1:69091-70008  NC_000001.10  \n",
       "3              NaN  chr1:160446-161525  NC_000001.10  \n",
       "4        NR_024368  chr1:317811-328581  NC_000001.10  \n",
       "...            ...                 ...           ...  \n",
       "2413168        NaN                 NaN           NaN  \n",
       "2413169        NaN                 NaN           NaN  \n",
       "2413170        NaN                 NaN           NaN  \n",
       "2413171        NaN                 NaN           NaN  \n",
       "2413172        NaN                 NaN           NaN  \n",
       "\n",
       "[2413173 rows x 8 columns]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'RANGE_STRAND', 'RANGE_START', 'RANGE_END', 'total_probes',\n",
       "       'GB_ACC', 'SPOT_ID', 'RANGE_GB'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression profiling of pediatric adrenocortical tumors of patients treated on the Children's Oncology Group XXX protocol.\"\n",
      "!Series_summary\t\"We have previously observed that expression of HLA genes associate with histology of adrenocortical tumors (PMID 17234769).\"\n",
      "!Series_summary\t\"Here, we used gene expression microarrays to associate the diagnostic tumor expression of these genes with outcome among 34 patients treated on the COG ARAR0332 protocol.\"\n",
      "!Series_overall_design\t\"We used microarrays to explore the expression profiles of a large group of uniformly-treated pediatric adrenocortical carcinomas.\"\n",
      "!Series_overall_design\t\"Specimens were harvested during surgery and snap frozen in liquid nitrogen to preserve tissue integrity.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1972883</th>\n",
       "      <th>GSM1972884</th>\n",
       "      <th>GSM1972885</th>\n",
       "      <th>GSM1972886</th>\n",
       "      <th>GSM1972887</th>\n",
       "      <th>GSM1972888</th>\n",
       "      <th>GSM1972889</th>\n",
       "      <th>GSM1972890</th>\n",
       "      <th>GSM1972891</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1972907</th>\n",
       "      <th>GSM1972908</th>\n",
       "      <th>GSM1972909</th>\n",
       "      <th>GSM1972910</th>\n",
       "      <th>GSM1972911</th>\n",
       "      <th>GSM1972912</th>\n",
       "      <th>GSM1972913</th>\n",
       "      <th>GSM1972914</th>\n",
       "      <th>GSM1972915</th>\n",
       "      <th>GSM1972916</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>...</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "      <td>histology: ACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Stage: III</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>Stage: III</td>\n",
       "      <td>Stage: II</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>Stage: II</td>\n",
       "      <td>Stage: IV</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>...</td>\n",
       "      <td>Stage: III</td>\n",
       "      <td>Stage: II</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>Stage: IV</td>\n",
       "      <td>Stage: III</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>Stage: I</td>\n",
       "      <td>Stage: IV</td>\n",
       "      <td>Stage: III</td>\n",
       "      <td>Stage: II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>efs.time: 5.07323750855578</td>\n",
       "      <td>efs.time: 5.17453798767967</td>\n",
       "      <td>efs.time: 4.33127994524298</td>\n",
       "      <td>efs.time: 4.50376454483231</td>\n",
       "      <td>efs.time: 4.29568788501027</td>\n",
       "      <td>efs.time: 5.48117727583847</td>\n",
       "      <td>efs.time: 4.290212183436</td>\n",
       "      <td>efs.time: 3.35112936344969</td>\n",
       "      <td>efs.time: 4.87063655030801</td>\n",
       "      <td>...</td>\n",
       "      <td>efs.time: 7.08829568788501</td>\n",
       "      <td>efs.time: 2.01232032854209</td>\n",
       "      <td>efs.time: 1.70841889117043</td>\n",
       "      <td>efs.time: 0.563997262149213</td>\n",
       "      <td>efs.time: 2.45311430527036</td>\n",
       "      <td>efs.time: 2.13004791238877</td>\n",
       "      <td>efs.time: 1.6290212183436</td>\n",
       "      <td>efs.time: 0.750171115674196</td>\n",
       "      <td>efs.time: 1.90828199863107</td>\n",
       "      <td>efs.time: 0.511978097193703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 1</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 1</td>\n",
       "      <td>efs.event: 0</td>\n",
       "      <td>efs.event: 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                  GSM1972883  \\\n",
       "0  !Sample_characteristics_ch1              histology: ACC   \n",
       "1  !Sample_characteristics_ch1                  Stage: III   \n",
       "2  !Sample_characteristics_ch1  efs.time: 5.07323750855578   \n",
       "3  !Sample_characteristics_ch1                efs.event: 0   \n",
       "\n",
       "                   GSM1972884                  GSM1972885  \\\n",
       "0              histology: ACC              histology: ACC   \n",
       "1                    Stage: I                    Stage: I   \n",
       "2  efs.time: 5.17453798767967  efs.time: 4.33127994524298   \n",
       "3                efs.event: 0                efs.event: 0   \n",
       "\n",
       "                   GSM1972886                  GSM1972887  \\\n",
       "0              histology: ACC              histology: ACC   \n",
       "1                  Stage: III                   Stage: II   \n",
       "2  efs.time: 4.50376454483231  efs.time: 4.29568788501027   \n",
       "3                efs.event: 0                efs.event: 0   \n",
       "\n",
       "                   GSM1972888                GSM1972889  \\\n",
       "0              histology: ACC            histology: ACC   \n",
       "1                    Stage: I                 Stage: II   \n",
       "2  efs.time: 5.48117727583847  efs.time: 4.290212183436   \n",
       "3                efs.event: 0              efs.event: 0   \n",
       "\n",
       "                   GSM1972890                  GSM1972891  ...  \\\n",
       "0              histology: ACC              histology: ACC  ...   \n",
       "1                   Stage: IV                    Stage: I  ...   \n",
       "2  efs.time: 3.35112936344969  efs.time: 4.87063655030801  ...   \n",
       "3                efs.event: 0                efs.event: 0  ...   \n",
       "\n",
       "                   GSM1972907                  GSM1972908  \\\n",
       "0              histology: ACC              histology: ACC   \n",
       "1                  Stage: III                   Stage: II   \n",
       "2  efs.time: 7.08829568788501  efs.time: 2.01232032854209   \n",
       "3                efs.event: 0                efs.event: 0   \n",
       "\n",
       "                   GSM1972909                   GSM1972910  \\\n",
       "0              histology: ACC               histology: ACC   \n",
       "1                    Stage: I                    Stage: IV   \n",
       "2  efs.time: 1.70841889117043  efs.time: 0.563997262149213   \n",
       "3                efs.event: 0                 efs.event: 1   \n",
       "\n",
       "                   GSM1972911                  GSM1972912  \\\n",
       "0              histology: ACC              histology: ACC   \n",
       "1                  Stage: III                    Stage: I   \n",
       "2  efs.time: 2.45311430527036  efs.time: 2.13004791238877   \n",
       "3                efs.event: 0                efs.event: 0   \n",
       "\n",
       "                  GSM1972913                   GSM1972914  \\\n",
       "0             histology: ACC               histology: ACC   \n",
       "1                   Stage: I                    Stage: IV   \n",
       "2  efs.time: 1.6290212183436  efs.time: 0.750171115674196   \n",
       "3               efs.event: 0                 efs.event: 1   \n",
       "\n",
       "                   GSM1972915                   GSM1972916  \n",
       "0              histology: ACC               histology: ACC  \n",
       "1                  Stage: III                    Stage: II  \n",
       "2  efs.time: 1.90828199863107  efs.time: 0.511978097193703  \n",
       "3                efs.event: 0                 efs.event: 1  \n",
       "\n",
       "[4 rows x 35 columns]"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE76019\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'histology: ACC'], dtype=object)"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'histology: ACC':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1972883</th>\n",
       "      <th>GSM1972884</th>\n",
       "      <th>GSM1972885</th>\n",
       "      <th>GSM1972886</th>\n",
       "      <th>GSM1972887</th>\n",
       "      <th>GSM1972888</th>\n",
       "      <th>GSM1972889</th>\n",
       "      <th>GSM1972890</th>\n",
       "      <th>GSM1972891</th>\n",
       "      <th>GSM1972892</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1972907</th>\n",
       "      <th>GSM1972908</th>\n",
       "      <th>GSM1972909</th>\n",
       "      <th>GSM1972910</th>\n",
       "      <th>GSM1972911</th>\n",
       "      <th>GSM1972912</th>\n",
       "      <th>GSM1972913</th>\n",
       "      <th>GSM1972914</th>\n",
       "      <th>GSM1972915</th>\n",
       "      <th>GSM1972916</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM1972883  GSM1972884  GSM1972885  GSM1972886  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM1972887  GSM1972888  GSM1972889  GSM1972890  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM1972891  GSM1972892  ...  GSM1972907  GSM1972908  \\\n",
       "Adrenocortical Cancer           1           1  ...           1           1   \n",
       "\n",
       "                       GSM1972909  GSM1972910  GSM1972911  GSM1972912  \\\n",
       "Adrenocortical Cancer           1           1           1           1   \n",
       "\n",
       "                       GSM1972913  GSM1972914  GSM1972915  GSM1972916  \n",
       "Adrenocortical Cancer           1           1           1           1  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1972883</th>\n",
       "      <th>GSM1972884</th>\n",
       "      <th>GSM1972885</th>\n",
       "      <th>GSM1972886</th>\n",
       "      <th>GSM1972887</th>\n",
       "      <th>GSM1972888</th>\n",
       "      <th>GSM1972889</th>\n",
       "      <th>GSM1972890</th>\n",
       "      <th>GSM1972891</th>\n",
       "      <th>GSM1972892</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1972907</th>\n",
       "      <th>GSM1972908</th>\n",
       "      <th>GSM1972909</th>\n",
       "      <th>GSM1972910</th>\n",
       "      <th>GSM1972911</th>\n",
       "      <th>GSM1972912</th>\n",
       "      <th>GSM1972913</th>\n",
       "      <th>GSM1972914</th>\n",
       "      <th>GSM1972915</th>\n",
       "      <th>GSM1972916</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_PM_s_at</th>\n",
       "      <td>11.198248</td>\n",
       "      <td>7.933991</td>\n",
       "      <td>7.960398</td>\n",
       "      <td>9.394266</td>\n",
       "      <td>10.052110</td>\n",
       "      <td>9.110888</td>\n",
       "      <td>10.281812</td>\n",
       "      <td>10.182810</td>\n",
       "      <td>7.545310</td>\n",
       "      <td>8.466753</td>\n",
       "      <td>...</td>\n",
       "      <td>10.139733</td>\n",
       "      <td>10.113103</td>\n",
       "      <td>7.980672</td>\n",
       "      <td>8.076911</td>\n",
       "      <td>9.057903</td>\n",
       "      <td>7.906223</td>\n",
       "      <td>9.819896</td>\n",
       "      <td>9.295794</td>\n",
       "      <td>9.881930</td>\n",
       "      <td>9.663820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_PM_at</th>\n",
       "      <td>6.636225</td>\n",
       "      <td>6.589175</td>\n",
       "      <td>7.139055</td>\n",
       "      <td>6.789730</td>\n",
       "      <td>6.894620</td>\n",
       "      <td>6.224652</td>\n",
       "      <td>6.027137</td>\n",
       "      <td>6.040128</td>\n",
       "      <td>7.009429</td>\n",
       "      <td>5.893735</td>\n",
       "      <td>...</td>\n",
       "      <td>5.476718</td>\n",
       "      <td>6.046398</td>\n",
       "      <td>6.339334</td>\n",
       "      <td>6.445096</td>\n",
       "      <td>7.831981</td>\n",
       "      <td>7.071057</td>\n",
       "      <td>6.239265</td>\n",
       "      <td>7.505965</td>\n",
       "      <td>6.833472</td>\n",
       "      <td>7.676311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_PM_at</th>\n",
       "      <td>3.567993</td>\n",
       "      <td>3.623477</td>\n",
       "      <td>4.011449</td>\n",
       "      <td>3.738795</td>\n",
       "      <td>3.371988</td>\n",
       "      <td>5.343177</td>\n",
       "      <td>3.733957</td>\n",
       "      <td>3.447104</td>\n",
       "      <td>4.302283</td>\n",
       "      <td>3.524911</td>\n",
       "      <td>...</td>\n",
       "      <td>3.187503</td>\n",
       "      <td>2.973845</td>\n",
       "      <td>2.946495</td>\n",
       "      <td>3.472866</td>\n",
       "      <td>6.236267</td>\n",
       "      <td>3.505690</td>\n",
       "      <td>3.362644</td>\n",
       "      <td>4.042302</td>\n",
       "      <td>3.317488</td>\n",
       "      <td>4.425911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_PM_at</th>\n",
       "      <td>5.908458</td>\n",
       "      <td>5.115113</td>\n",
       "      <td>6.658453</td>\n",
       "      <td>5.386049</td>\n",
       "      <td>5.481440</td>\n",
       "      <td>5.187689</td>\n",
       "      <td>5.126747</td>\n",
       "      <td>6.048818</td>\n",
       "      <td>6.691498</td>\n",
       "      <td>5.234772</td>\n",
       "      <td>...</td>\n",
       "      <td>6.683220</td>\n",
       "      <td>5.072790</td>\n",
       "      <td>6.936309</td>\n",
       "      <td>6.305064</td>\n",
       "      <td>4.728112</td>\n",
       "      <td>4.960299</td>\n",
       "      <td>6.832099</td>\n",
       "      <td>5.184749</td>\n",
       "      <td>5.206011</td>\n",
       "      <td>5.279768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_PM_g_at</th>\n",
       "      <td>2.419798</td>\n",
       "      <td>2.585927</td>\n",
       "      <td>2.580811</td>\n",
       "      <td>2.574198</td>\n",
       "      <td>2.587390</td>\n",
       "      <td>2.546555</td>\n",
       "      <td>2.569797</td>\n",
       "      <td>2.546564</td>\n",
       "      <td>2.604273</td>\n",
       "      <td>2.546547</td>\n",
       "      <td>...</td>\n",
       "      <td>2.418516</td>\n",
       "      <td>2.644498</td>\n",
       "      <td>2.739033</td>\n",
       "      <td>2.756333</td>\n",
       "      <td>2.813312</td>\n",
       "      <td>2.584168</td>\n",
       "      <td>2.686655</td>\n",
       "      <td>2.524916</td>\n",
       "      <td>2.545832</td>\n",
       "      <td>2.671332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>5.841242</td>\n",
       "      <td>5.736721</td>\n",
       "      <td>5.930751</td>\n",
       "      <td>5.975872</td>\n",
       "      <td>5.439575</td>\n",
       "      <td>5.470157</td>\n",
       "      <td>4.898815</td>\n",
       "      <td>6.075566</td>\n",
       "      <td>5.827021</td>\n",
       "      <td>6.247272</td>\n",
       "      <td>...</td>\n",
       "      <td>6.464635</td>\n",
       "      <td>6.477860</td>\n",
       "      <td>5.318148</td>\n",
       "      <td>5.011828</td>\n",
       "      <td>5.184427</td>\n",
       "      <td>5.472901</td>\n",
       "      <td>5.296042</td>\n",
       "      <td>6.166961</td>\n",
       "      <td>5.266876</td>\n",
       "      <td>5.218927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>6.882192</td>\n",
       "      <td>6.936577</td>\n",
       "      <td>6.894621</td>\n",
       "      <td>7.019251</td>\n",
       "      <td>6.407738</td>\n",
       "      <td>6.998752</td>\n",
       "      <td>5.947718</td>\n",
       "      <td>6.902733</td>\n",
       "      <td>6.820512</td>\n",
       "      <td>7.359015</td>\n",
       "      <td>...</td>\n",
       "      <td>7.446475</td>\n",
       "      <td>7.737522</td>\n",
       "      <td>6.722462</td>\n",
       "      <td>6.484635</td>\n",
       "      <td>6.614603</td>\n",
       "      <td>6.902994</td>\n",
       "      <td>6.762439</td>\n",
       "      <td>7.857219</td>\n",
       "      <td>6.839717</td>\n",
       "      <td>6.623335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>2.554189</td>\n",
       "      <td>2.572302</td>\n",
       "      <td>2.573867</td>\n",
       "      <td>2.538020</td>\n",
       "      <td>2.481296</td>\n",
       "      <td>2.700988</td>\n",
       "      <td>2.531195</td>\n",
       "      <td>2.713880</td>\n",
       "      <td>2.710393</td>\n",
       "      <td>2.684654</td>\n",
       "      <td>...</td>\n",
       "      <td>2.604511</td>\n",
       "      <td>2.563804</td>\n",
       "      <td>2.426904</td>\n",
       "      <td>2.620220</td>\n",
       "      <td>2.485755</td>\n",
       "      <td>2.644209</td>\n",
       "      <td>2.569908</td>\n",
       "      <td>2.580324</td>\n",
       "      <td>2.593163</td>\n",
       "      <td>2.400546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>2.728991</td>\n",
       "      <td>2.746145</td>\n",
       "      <td>2.812163</td>\n",
       "      <td>2.769375</td>\n",
       "      <td>2.712377</td>\n",
       "      <td>2.834052</td>\n",
       "      <td>2.820497</td>\n",
       "      <td>3.123452</td>\n",
       "      <td>2.807358</td>\n",
       "      <td>2.787314</td>\n",
       "      <td>...</td>\n",
       "      <td>2.722211</td>\n",
       "      <td>2.884076</td>\n",
       "      <td>2.907535</td>\n",
       "      <td>2.882485</td>\n",
       "      <td>2.802728</td>\n",
       "      <td>2.752660</td>\n",
       "      <td>2.800514</td>\n",
       "      <td>2.869417</td>\n",
       "      <td>2.901859</td>\n",
       "      <td>2.733121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>2.812298</td>\n",
       "      <td>2.793757</td>\n",
       "      <td>2.809920</td>\n",
       "      <td>2.799200</td>\n",
       "      <td>2.742040</td>\n",
       "      <td>2.723134</td>\n",
       "      <td>2.575061</td>\n",
       "      <td>2.969765</td>\n",
       "      <td>2.890798</td>\n",
       "      <td>2.762177</td>\n",
       "      <td>...</td>\n",
       "      <td>2.823875</td>\n",
       "      <td>2.710125</td>\n",
       "      <td>2.799325</td>\n",
       "      <td>2.834195</td>\n",
       "      <td>2.925038</td>\n",
       "      <td>2.676069</td>\n",
       "      <td>2.775004</td>\n",
       "      <td>2.757608</td>\n",
       "      <td>2.747517</td>\n",
       "      <td>2.861152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54715 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM1972883  GSM1972884  GSM1972885  GSM1972886  GSM1972887  \\\n",
       "ID                                                                            \n",
       "1007_PM_s_at      11.198248    7.933991    7.960398    9.394266   10.052110   \n",
       "1053_PM_at         6.636225    6.589175    7.139055    6.789730    6.894620   \n",
       "117_PM_at          3.567993    3.623477    4.011449    3.738795    3.371988   \n",
       "121_PM_at          5.908458    5.115113    6.658453    5.386049    5.481440   \n",
       "1255_PM_g_at       2.419798    2.585927    2.580811    2.574198    2.587390   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     5.841242    5.736721    5.930751    5.975872    5.439575   \n",
       "AFFX-ThrX-M_at     6.882192    6.936577    6.894621    7.019251    6.407738   \n",
       "AFFX-TrpnX-3_at    2.554189    2.572302    2.573867    2.538020    2.481296   \n",
       "AFFX-TrpnX-5_at    2.728991    2.746145    2.812163    2.769375    2.712377   \n",
       "AFFX-TrpnX-M_at    2.812298    2.793757    2.809920    2.799200    2.742040   \n",
       "\n",
       "                 GSM1972888  GSM1972889  GSM1972890  GSM1972891  GSM1972892  \\\n",
       "ID                                                                            \n",
       "1007_PM_s_at       9.110888   10.281812   10.182810    7.545310    8.466753   \n",
       "1053_PM_at         6.224652    6.027137    6.040128    7.009429    5.893735   \n",
       "117_PM_at          5.343177    3.733957    3.447104    4.302283    3.524911   \n",
       "121_PM_at          5.187689    5.126747    6.048818    6.691498    5.234772   \n",
       "1255_PM_g_at       2.546555    2.569797    2.546564    2.604273    2.546547   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     5.470157    4.898815    6.075566    5.827021    6.247272   \n",
       "AFFX-ThrX-M_at     6.998752    5.947718    6.902733    6.820512    7.359015   \n",
       "AFFX-TrpnX-3_at    2.700988    2.531195    2.713880    2.710393    2.684654   \n",
       "AFFX-TrpnX-5_at    2.834052    2.820497    3.123452    2.807358    2.787314   \n",
       "AFFX-TrpnX-M_at    2.723134    2.575061    2.969765    2.890798    2.762177   \n",
       "\n",
       "                 ...  GSM1972907  GSM1972908  GSM1972909  GSM1972910  \\\n",
       "ID               ...                                                   \n",
       "1007_PM_s_at     ...   10.139733   10.113103    7.980672    8.076911   \n",
       "1053_PM_at       ...    5.476718    6.046398    6.339334    6.445096   \n",
       "117_PM_at        ...    3.187503    2.973845    2.946495    3.472866   \n",
       "121_PM_at        ...    6.683220    5.072790    6.936309    6.305064   \n",
       "1255_PM_g_at     ...    2.418516    2.644498    2.739033    2.756333   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at   ...    6.464635    6.477860    5.318148    5.011828   \n",
       "AFFX-ThrX-M_at   ...    7.446475    7.737522    6.722462    6.484635   \n",
       "AFFX-TrpnX-3_at  ...    2.604511    2.563804    2.426904    2.620220   \n",
       "AFFX-TrpnX-5_at  ...    2.722211    2.884076    2.907535    2.882485   \n",
       "AFFX-TrpnX-M_at  ...    2.823875    2.710125    2.799325    2.834195   \n",
       "\n",
       "                 GSM1972911  GSM1972912  GSM1972913  GSM1972914  GSM1972915  \\\n",
       "ID                                                                            \n",
       "1007_PM_s_at       9.057903    7.906223    9.819896    9.295794    9.881930   \n",
       "1053_PM_at         7.831981    7.071057    6.239265    7.505965    6.833472   \n",
       "117_PM_at          6.236267    3.505690    3.362644    4.042302    3.317488   \n",
       "121_PM_at          4.728112    4.960299    6.832099    5.184749    5.206011   \n",
       "1255_PM_g_at       2.813312    2.584168    2.686655    2.524916    2.545832   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     5.184427    5.472901    5.296042    6.166961    5.266876   \n",
       "AFFX-ThrX-M_at     6.614603    6.902994    6.762439    7.857219    6.839717   \n",
       "AFFX-TrpnX-3_at    2.485755    2.644209    2.569908    2.580324    2.593163   \n",
       "AFFX-TrpnX-5_at    2.802728    2.752660    2.800514    2.869417    2.901859   \n",
       "AFFX-TrpnX-M_at    2.925038    2.676069    2.775004    2.757608    2.747517   \n",
       "\n",
       "                 GSM1972916  \n",
       "ID                           \n",
       "1007_PM_s_at       9.663820  \n",
       "1053_PM_at         7.676311  \n",
       "117_PM_at          4.425911  \n",
       "121_PM_at          5.279768  \n",
       "1255_PM_g_at       2.671332  \n",
       "...                     ...  \n",
       "AFFX-ThrX-5_at     5.218927  \n",
       "AFFX-ThrX-M_at     6.623335  \n",
       "AFFX-TrpnX-3_at    2.400546  \n",
       "AFFX-TrpnX-5_at    2.733121  \n",
       "AFFX-TrpnX-M_at    2.861152  \n",
       "\n",
       "[54715 rows x 34 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_PM_s_at', '1053_PM_at', '117_PM_at', '121_PM_at', '1255_PM_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Aug 20, 2010', 'Aug 20, 2010', 'Aug 20, 2010', 'Aug 20, 2010', 'Aug 20, 2010'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001954 /// NM_013993 /// NM_013994', 'NM_002914 /// NM_181471', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein amino acid phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0031100 // organ regeneration // inferred from electronic annotation /// 0043583 // ear development // inferred from electronic annotation /// 0043588 // skin development // inferred from electronic annotation /// 0051789 // response to protein stimulus // inferred from electronic annotation /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation', '0006260 // DNA replication // not recorded /// 0006260 // DNA replication // inferred from electronic annotation /// 0006297 // nucleotide-excision repair, DNA gap filling // not recorded /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation', '0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement', '0001656 // metanephros development // inferred from electronic annotation /// 0006350 // transcription // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from electronic annotation /// 0045449 // regulation of transcription // inferred from electronic annotation /// 0045893 // positive regulation of transcription, DNA-dependent // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-dependent // inferred from direct assay /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from electronic annotation', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007601 // visual perception // traceable author statement /// 0007602 // phototransduction // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from electronic annotation /// 0005887 // integral to plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral to membrane // inferred from electronic annotation /// 0016323 // basolateral plasma membrane // inferred from electronic annotation', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // not recorded /// 0005663 // DNA replication factor C complex // inferred from direct assay /// 0005663 // DNA replication factor C complex // inferred from electronic annotation', nan, '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005654 // nucleoplasm // inferred from electronic annotation', '0016020 // membrane // inferred from electronic annotation'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0004872 // receptor activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005515 // protein binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0003689 // DNA clamp loader activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0005524 // ATP binding // traceable author statement /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation', '0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from electronic annotation /// 0003700 // transcription factor activity // traceable author statement /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from sequence or structural similarity /// 0005515 // protein binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0016563 // transcription activator activity // inferred from sequence or structural similarity /// 0016563 // transcription activator activity // inferred from direct assay /// 0016563 // transcription activator activity // inferred from electronic annotation /// 0043565 // sequence-specific DNA binding // inferred from electronic annotation', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // traceable author statement /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_PM_s_at', '1053_PM_at', '117_PM_at', '121_PM_at', '1255_PM_g_at', '1294_PM_at', '1316_PM_at', '1320_PM_at', '1405_PM_i_at', '1431_PM_at', '1438_PM_at', '1487_PM_at', '1494_PM_f_at', '1552256_PM_a_at', '1552257_PM_a_at', '1552258_PM_at', '1552261_PM_at', '1552263_PM_at', '1552264_PM_a_at', '1552266_PM_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_PM_s_at', '1053_PM_at', '117_PM_at', '121_PM_at', '1255_PM_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Aug 20, 2010', 'Aug 20, 2010', 'Aug 20, 2010', 'Aug 20, 2010', 'Aug 20, 2010'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001954 /// NM_013993 /// NM_013994', 'NM_002914 /// NM_181471', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein amino acid phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0031100 // organ regeneration // inferred from electronic annotation /// 0043583 // ear development // inferred from electronic annotation /// 0043588 // skin development // inferred from electronic annotation /// 0051789 // response to protein stimulus // inferred from electronic annotation /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation', '0006260 // DNA replication // not recorded /// 0006260 // DNA replication // inferred from electronic annotation /// 0006297 // nucleotide-excision repair, DNA gap filling // not recorded /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation', '0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement', '0001656 // metanephros development // inferred from electronic annotation /// 0006350 // transcription // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from electronic annotation /// 0045449 // regulation of transcription // inferred from electronic annotation /// 0045893 // positive regulation of transcription, DNA-dependent // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-dependent // inferred from direct assay /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from electronic annotation', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007601 // visual perception // traceable author statement /// 0007602 // phototransduction // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from electronic annotation /// 0005887 // integral to plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral to membrane // inferred from electronic annotation /// 0016323 // basolateral plasma membrane // inferred from electronic annotation', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // not recorded /// 0005663 // DNA replication factor C complex // inferred from direct assay /// 0005663 // DNA replication factor C complex // inferred from electronic annotation', nan, '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005654 // nucleoplasm // inferred from electronic annotation', '0016020 // membrane // inferred from electronic annotation'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0004872 // receptor activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005515 // protein binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0003689 // DNA clamp loader activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0005524 // ATP binding // traceable author statement /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation', '0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from electronic annotation /// 0003700 // transcription factor activity // traceable author statement /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from sequence or structural similarity /// 0005515 // protein binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0016563 // transcription activator activity // inferred from sequence or structural similarity /// 0016563 // transcription activator activity // inferred from direct assay /// 0016563 // transcription activator activity // inferred from electronic annotation /// 0043565 // sequence-specific DNA binding // inferred from electronic annotation', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // traceable author statement /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_PM_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Aug 20, 2010</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1</td>\n",
       "      <td>DDR1</td>\n",
       "      <td>780</td>\n",
       "      <td>NM_001954 /// NM_013993 /// NM_013994</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_PM_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Aug 20, 2010</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_002914 /// NM_181471</td>\n",
       "      <td>0006260 // DNA replication // not recorded ///...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_PM_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Aug 20, 2010</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0006950 // response to stress // inferred from...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_PM_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Aug 20, 2010</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001656 // metanephros development // inferred...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0003677 // DNA binding // inferred from direct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_PM_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Aug 20, 2010</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0016020 // membrane // inferred from electroni...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915054</th>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>5.218926936</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915055</th>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>6.623334728</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915056</th>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>2.400545959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915057</th>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>2.733121104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915058</th>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>2.861152041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1915059 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID       GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0           1007_PM_s_at       U48705     NaN            Homo sapiens   \n",
       "1             1053_PM_at       M87338     NaN            Homo sapiens   \n",
       "2              117_PM_at       X51757     NaN            Homo sapiens   \n",
       "3              121_PM_at       X69699     NaN            Homo sapiens   \n",
       "4           1255_PM_g_at       L36861     NaN            Homo sapiens   \n",
       "...                  ...          ...     ...                     ...   \n",
       "1915054   AFFX-ThrX-5_at  5.218926936     NaN                     NaN   \n",
       "1915055   AFFX-ThrX-M_at  6.623334728     NaN                     NaN   \n",
       "1915056  AFFX-TrpnX-3_at  2.400545959     NaN                     NaN   \n",
       "1915057  AFFX-TrpnX-5_at  2.733121104     NaN                     NaN   \n",
       "1915058  AFFX-TrpnX-M_at  2.861152041     NaN                     NaN   \n",
       "\n",
       "        Annotation Date      Sequence Type                  Sequence Source  \\\n",
       "0          Aug 20, 2010  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1          Aug 20, 2010  Exemplar sequence                          GenBank   \n",
       "2          Aug 20, 2010  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3          Aug 20, 2010  Exemplar sequence                          GenBank   \n",
       "4          Aug 20, 2010  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                 ...                ...                              ...   \n",
       "1915054             NaN                NaN                              NaN   \n",
       "1915055             NaN                NaN                              NaN   \n",
       "1915056             NaN                NaN                              NaN   \n",
       "1915057             NaN                NaN                              NaN   \n",
       "1915058             NaN                NaN                              NaN   \n",
       "\n",
       "                                        Target Description  \\\n",
       "0        U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1        M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2        X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3        X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4        L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                    ...   \n",
       "1915054                                                NaN   \n",
       "1915055                                                NaN   \n",
       "1915056                                                NaN   \n",
       "1915057                                                NaN   \n",
       "1915058                                                NaN   \n",
       "\n",
       "        Representative Public ID                                   Gene Title  \\\n",
       "0                         U48705  discoidin domain receptor tyrosine kinase 1   \n",
       "1                         M87338  replication factor C (activator 1) 2, 40kDa   \n",
       "2                         X51757         heat shock 70kDa protein 6 (HSP70B')   \n",
       "3                         X69699                                 paired box 8   \n",
       "4                         L36861      guanylate cyclase activator 1A (retina)   \n",
       "...                          ...                                          ...   \n",
       "1915054                      NaN                                          NaN   \n",
       "1915055                      NaN                                          NaN   \n",
       "1915056                      NaN                                          NaN   \n",
       "1915057                      NaN                                          NaN   \n",
       "1915058                      NaN                                          NaN   \n",
       "\n",
       "        Gene Symbol ENTREZ_GENE_ID  \\\n",
       "0              DDR1            780   \n",
       "1              RFC2           5982   \n",
       "2             HSPA6           3310   \n",
       "3              PAX8           7849   \n",
       "4            GUCA1A           2978   \n",
       "...             ...            ...   \n",
       "1915054         NaN            NaN   \n",
       "1915055         NaN            NaN   \n",
       "1915056         NaN            NaN   \n",
       "1915057         NaN            NaN   \n",
       "1915058         NaN            NaN   \n",
       "\n",
       "                                      RefSeq Transcript ID  \\\n",
       "0                    NM_001954 /// NM_013993 /// NM_013994   \n",
       "1                                  NM_002914 /// NM_181471   \n",
       "2                                                NM_002155   \n",
       "3        NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                                                NM_000409   \n",
       "...                                                    ...   \n",
       "1915054                                                NaN   \n",
       "1915055                                                NaN   \n",
       "1915056                                                NaN   \n",
       "1915057                                                NaN   \n",
       "1915058                                                NaN   \n",
       "\n",
       "                          Gene Ontology Biological Process  \\\n",
       "0        0001558 // regulation of cell growth // inferr...   \n",
       "1        0006260 // DNA replication // not recorded ///...   \n",
       "2        0006950 // response to stress // inferred from...   \n",
       "3        0001656 // metanephros development // inferred...   \n",
       "4        0007165 // signal transduction // non-traceabl...   \n",
       "...                                                    ...   \n",
       "1915054                                                NaN   \n",
       "1915055                                                NaN   \n",
       "1915056                                                NaN   \n",
       "1915057                                                NaN   \n",
       "1915058                                                NaN   \n",
       "\n",
       "                          Gene Ontology Cellular Component  \\\n",
       "0        0005576 // extracellular region // inferred fr...   \n",
       "1        0005634 // nucleus // inferred from electronic...   \n",
       "2                                                      NaN   \n",
       "3        0005634 // nucleus // inferred from electronic...   \n",
       "4        0016020 // membrane // inferred from electroni...   \n",
       "...                                                    ...   \n",
       "1915054                                                NaN   \n",
       "1915055                                                NaN   \n",
       "1915056                                                NaN   \n",
       "1915057                                                NaN   \n",
       "1915058                                                NaN   \n",
       "\n",
       "                          Gene Ontology Molecular Function  \n",
       "0        0000166 // nucleotide binding // inferred from...  \n",
       "1        0000166 // nucleotide binding // inferred from...  \n",
       "2        0000166 // nucleotide binding // inferred from...  \n",
       "3        0003677 // DNA binding // inferred from direct...  \n",
       "4        0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                    ...  \n",
       "1915054                                                NaN  \n",
       "1915055                                                NaN  \n",
       "1915056                                                NaN  \n",
       "1915057                                                NaN  \n",
       "1915058                                                NaN  \n",
       "\n",
       "[1915059 rows x 16 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1972883</th>\n",
       "      <th>GSM1972884</th>\n",
       "      <th>GSM1972885</th>\n",
       "      <th>GSM1972886</th>\n",
       "      <th>GSM1972887</th>\n",
       "      <th>GSM1972888</th>\n",
       "      <th>GSM1972889</th>\n",
       "      <th>GSM1972890</th>\n",
       "      <th>GSM1972891</th>\n",
       "      <th>GSM1972892</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1972907</th>\n",
       "      <th>GSM1972908</th>\n",
       "      <th>GSM1972909</th>\n",
       "      <th>GSM1972910</th>\n",
       "      <th>GSM1972911</th>\n",
       "      <th>GSM1972912</th>\n",
       "      <th>GSM1972913</th>\n",
       "      <th>GSM1972914</th>\n",
       "      <th>GSM1972915</th>\n",
       "      <th>GSM1972916</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>8.146158</td>\n",
       "      <td>10.977810</td>\n",
       "      <td>10.817974</td>\n",
       "      <td>8.328435</td>\n",
       "      <td>7.702278</td>\n",
       "      <td>9.928567</td>\n",
       "      <td>10.189140</td>\n",
       "      <td>10.418385</td>\n",
       "      <td>10.703676</td>\n",
       "      <td>6.957320</td>\n",
       "      <td>...</td>\n",
       "      <td>11.107936</td>\n",
       "      <td>7.462309</td>\n",
       "      <td>10.275823</td>\n",
       "      <td>10.943833</td>\n",
       "      <td>9.631248</td>\n",
       "      <td>9.684394</td>\n",
       "      <td>8.455126</td>\n",
       "      <td>8.191178</td>\n",
       "      <td>6.759635</td>\n",
       "      <td>6.429762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>7.723734</td>\n",
       "      <td>8.084656</td>\n",
       "      <td>6.997038</td>\n",
       "      <td>6.348997</td>\n",
       "      <td>7.657459</td>\n",
       "      <td>5.774166</td>\n",
       "      <td>6.133703</td>\n",
       "      <td>4.234780</td>\n",
       "      <td>5.267539</td>\n",
       "      <td>5.438076</td>\n",
       "      <td>...</td>\n",
       "      <td>7.690533</td>\n",
       "      <td>5.694893</td>\n",
       "      <td>8.007111</td>\n",
       "      <td>6.423638</td>\n",
       "      <td>5.311714</td>\n",
       "      <td>6.375171</td>\n",
       "      <td>7.027140</td>\n",
       "      <td>7.036405</td>\n",
       "      <td>8.993887</td>\n",
       "      <td>7.220674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACOT2</th>\n",
       "      <td>8.106406</td>\n",
       "      <td>7.817518</td>\n",
       "      <td>8.597533</td>\n",
       "      <td>7.436279</td>\n",
       "      <td>7.458640</td>\n",
       "      <td>6.967079</td>\n",
       "      <td>7.762308</td>\n",
       "      <td>7.919932</td>\n",
       "      <td>7.028197</td>\n",
       "      <td>8.076595</td>\n",
       "      <td>...</td>\n",
       "      <td>7.144096</td>\n",
       "      <td>8.266708</td>\n",
       "      <td>7.390831</td>\n",
       "      <td>8.582612</td>\n",
       "      <td>8.178542</td>\n",
       "      <td>7.883154</td>\n",
       "      <td>8.806078</td>\n",
       "      <td>7.615123</td>\n",
       "      <td>6.970443</td>\n",
       "      <td>7.752594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM2B</th>\n",
       "      <td>2.661260</td>\n",
       "      <td>2.671672</td>\n",
       "      <td>2.593285</td>\n",
       "      <td>2.577828</td>\n",
       "      <td>2.405197</td>\n",
       "      <td>2.640131</td>\n",
       "      <td>2.487925</td>\n",
       "      <td>2.582874</td>\n",
       "      <td>2.530571</td>\n",
       "      <td>2.657053</td>\n",
       "      <td>...</td>\n",
       "      <td>2.613758</td>\n",
       "      <td>2.626113</td>\n",
       "      <td>2.800945</td>\n",
       "      <td>2.511766</td>\n",
       "      <td>2.525154</td>\n",
       "      <td>2.578395</td>\n",
       "      <td>2.667927</td>\n",
       "      <td>2.626038</td>\n",
       "      <td>2.625373</td>\n",
       "      <td>2.757021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACSM2B</th>\n",
       "      <td>2.790370</td>\n",
       "      <td>2.759889</td>\n",
       "      <td>2.838374</td>\n",
       "      <td>2.866704</td>\n",
       "      <td>2.899911</td>\n",
       "      <td>2.740075</td>\n",
       "      <td>2.764121</td>\n",
       "      <td>2.775007</td>\n",
       "      <td>2.780009</td>\n",
       "      <td>3.104381</td>\n",
       "      <td>...</td>\n",
       "      <td>2.966712</td>\n",
       "      <td>2.824844</td>\n",
       "      <td>2.741376</td>\n",
       "      <td>2.803511</td>\n",
       "      <td>2.991721</td>\n",
       "      <td>2.802611</td>\n",
       "      <td>2.969826</td>\n",
       "      <td>2.868441</td>\n",
       "      <td>2.845737</td>\n",
       "      <td>2.812486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>3.707414</td>\n",
       "      <td>3.824522</td>\n",
       "      <td>3.970674</td>\n",
       "      <td>3.953966</td>\n",
       "      <td>3.314021</td>\n",
       "      <td>3.750447</td>\n",
       "      <td>4.481623</td>\n",
       "      <td>5.350892</td>\n",
       "      <td>4.075008</td>\n",
       "      <td>4.400005</td>\n",
       "      <td>...</td>\n",
       "      <td>3.663827</td>\n",
       "      <td>3.892046</td>\n",
       "      <td>3.643871</td>\n",
       "      <td>3.635140</td>\n",
       "      <td>2.974060</td>\n",
       "      <td>3.826820</td>\n",
       "      <td>4.372133</td>\n",
       "      <td>3.554603</td>\n",
       "      <td>3.158875</td>\n",
       "      <td>3.565214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>4.953882</td>\n",
       "      <td>4.852440</td>\n",
       "      <td>4.961707</td>\n",
       "      <td>5.035332</td>\n",
       "      <td>5.086926</td>\n",
       "      <td>5.210269</td>\n",
       "      <td>5.084559</td>\n",
       "      <td>5.412141</td>\n",
       "      <td>4.688110</td>\n",
       "      <td>5.128178</td>\n",
       "      <td>...</td>\n",
       "      <td>5.296990</td>\n",
       "      <td>4.951781</td>\n",
       "      <td>4.795016</td>\n",
       "      <td>4.912349</td>\n",
       "      <td>5.306210</td>\n",
       "      <td>4.846485</td>\n",
       "      <td>5.563509</td>\n",
       "      <td>4.626624</td>\n",
       "      <td>5.128579</td>\n",
       "      <td>4.568179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>8.010302</td>\n",
       "      <td>7.458704</td>\n",
       "      <td>7.640166</td>\n",
       "      <td>7.455243</td>\n",
       "      <td>7.439838</td>\n",
       "      <td>8.055964</td>\n",
       "      <td>7.422100</td>\n",
       "      <td>7.491325</td>\n",
       "      <td>7.031681</td>\n",
       "      <td>7.838766</td>\n",
       "      <td>...</td>\n",
       "      <td>7.369171</td>\n",
       "      <td>7.530654</td>\n",
       "      <td>7.545323</td>\n",
       "      <td>7.729064</td>\n",
       "      <td>7.862185</td>\n",
       "      <td>7.626514</td>\n",
       "      <td>7.703143</td>\n",
       "      <td>7.479374</td>\n",
       "      <td>7.334361</td>\n",
       "      <td>7.149404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psiTPTE22</th>\n",
       "      <td>4.675962</td>\n",
       "      <td>3.057613</td>\n",
       "      <td>4.286915</td>\n",
       "      <td>2.946979</td>\n",
       "      <td>5.315133</td>\n",
       "      <td>3.884333</td>\n",
       "      <td>3.142242</td>\n",
       "      <td>2.875372</td>\n",
       "      <td>3.487803</td>\n",
       "      <td>2.855120</td>\n",
       "      <td>...</td>\n",
       "      <td>2.890170</td>\n",
       "      <td>2.886104</td>\n",
       "      <td>3.314108</td>\n",
       "      <td>4.380311</td>\n",
       "      <td>3.229161</td>\n",
       "      <td>3.017693</td>\n",
       "      <td>3.086272</td>\n",
       "      <td>2.787978</td>\n",
       "      <td>2.905757</td>\n",
       "      <td>5.173520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tAKR</th>\n",
       "      <td>2.751727</td>\n",
       "      <td>3.113977</td>\n",
       "      <td>2.869546</td>\n",
       "      <td>2.441583</td>\n",
       "      <td>2.660435</td>\n",
       "      <td>3.063617</td>\n",
       "      <td>2.774778</td>\n",
       "      <td>3.070187</td>\n",
       "      <td>2.857389</td>\n",
       "      <td>2.788373</td>\n",
       "      <td>...</td>\n",
       "      <td>2.957765</td>\n",
       "      <td>2.786444</td>\n",
       "      <td>2.417844</td>\n",
       "      <td>2.891852</td>\n",
       "      <td>2.619512</td>\n",
       "      <td>2.898136</td>\n",
       "      <td>2.759239</td>\n",
       "      <td>2.697102</td>\n",
       "      <td>2.797410</td>\n",
       "      <td>2.793014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21784 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM1972883  GSM1972884  GSM1972885  GSM1972886  GSM1972887  \\\n",
       "Gene                                                                    \n",
       " ABCB4       8.146158   10.977810   10.817974    8.328435    7.702278   \n",
       " ABCC6P2     7.723734    8.084656    6.997038    6.348997    7.657459   \n",
       " ACOT2       8.106406    7.817518    8.597533    7.436279    7.458640   \n",
       " ACSM2B      2.661260    2.671672    2.593285    2.577828    2.405197   \n",
       " ACSM2B      2.790370    2.759889    2.838374    2.866704    2.899911   \n",
       "...               ...         ...         ...         ...         ...   \n",
       "ZYX          3.707414    3.824522    3.970674    3.953966    3.314021   \n",
       "ZZEF1        4.953882    4.852440    4.961707    5.035332    5.086926   \n",
       "ZZZ3         8.010302    7.458704    7.640166    7.455243    7.439838   \n",
       "psiTPTE22    4.675962    3.057613    4.286915    2.946979    5.315133   \n",
       "tAKR         2.751727    3.113977    2.869546    2.441583    2.660435   \n",
       "\n",
       "           GSM1972888  GSM1972889  GSM1972890  GSM1972891  GSM1972892  ...  \\\n",
       "Gene                                                                   ...   \n",
       " ABCB4       9.928567   10.189140   10.418385   10.703676    6.957320  ...   \n",
       " ABCC6P2     5.774166    6.133703    4.234780    5.267539    5.438076  ...   \n",
       " ACOT2       6.967079    7.762308    7.919932    7.028197    8.076595  ...   \n",
       " ACSM2B      2.640131    2.487925    2.582874    2.530571    2.657053  ...   \n",
       " ACSM2B      2.740075    2.764121    2.775007    2.780009    3.104381  ...   \n",
       "...               ...         ...         ...         ...         ...  ...   \n",
       "ZYX          3.750447    4.481623    5.350892    4.075008    4.400005  ...   \n",
       "ZZEF1        5.210269    5.084559    5.412141    4.688110    5.128178  ...   \n",
       "ZZZ3         8.055964    7.422100    7.491325    7.031681    7.838766  ...   \n",
       "psiTPTE22    3.884333    3.142242    2.875372    3.487803    2.855120  ...   \n",
       "tAKR         3.063617    2.774778    3.070187    2.857389    2.788373  ...   \n",
       "\n",
       "           GSM1972907  GSM1972908  GSM1972909  GSM1972910  GSM1972911  \\\n",
       "Gene                                                                    \n",
       " ABCB4      11.107936    7.462309   10.275823   10.943833    9.631248   \n",
       " ABCC6P2     7.690533    5.694893    8.007111    6.423638    5.311714   \n",
       " ACOT2       7.144096    8.266708    7.390831    8.582612    8.178542   \n",
       " ACSM2B      2.613758    2.626113    2.800945    2.511766    2.525154   \n",
       " ACSM2B      2.966712    2.824844    2.741376    2.803511    2.991721   \n",
       "...               ...         ...         ...         ...         ...   \n",
       "ZYX          3.663827    3.892046    3.643871    3.635140    2.974060   \n",
       "ZZEF1        5.296990    4.951781    4.795016    4.912349    5.306210   \n",
       "ZZZ3         7.369171    7.530654    7.545323    7.729064    7.862185   \n",
       "psiTPTE22    2.890170    2.886104    3.314108    4.380311    3.229161   \n",
       "tAKR         2.957765    2.786444    2.417844    2.891852    2.619512   \n",
       "\n",
       "           GSM1972912  GSM1972913  GSM1972914  GSM1972915  GSM1972916  \n",
       "Gene                                                                   \n",
       " ABCB4       9.684394    8.455126    8.191178    6.759635    6.429762  \n",
       " ABCC6P2     6.375171    7.027140    7.036405    8.993887    7.220674  \n",
       " ACOT2       7.883154    8.806078    7.615123    6.970443    7.752594  \n",
       " ACSM2B      2.578395    2.667927    2.626038    2.625373    2.757021  \n",
       " ACSM2B      2.802611    2.969826    2.868441    2.845737    2.812486  \n",
       "...               ...         ...         ...         ...         ...  \n",
       "ZYX          3.826820    4.372133    3.554603    3.158875    3.565214  \n",
       "ZZEF1        4.846485    5.563509    4.626624    5.128579    4.568179  \n",
       "ZZZ3         7.626514    7.703143    7.479374    7.334361    7.149404  \n",
       "psiTPTE22    3.017693    3.086272    2.787978    2.905757    5.173520  \n",
       "tAKR         2.898136    2.759239    2.697102    2.797410    2.793014  \n",
       "\n",
       "[21784 rows x 34 columns]"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1972883</th>\n",
       "      <th>GSM1972884</th>\n",
       "      <th>GSM1972885</th>\n",
       "      <th>GSM1972886</th>\n",
       "      <th>GSM1972887</th>\n",
       "      <th>GSM1972888</th>\n",
       "      <th>GSM1972889</th>\n",
       "      <th>GSM1972890</th>\n",
       "      <th>GSM1972891</th>\n",
       "      <th>GSM1972892</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1972907</th>\n",
       "      <th>GSM1972908</th>\n",
       "      <th>GSM1972909</th>\n",
       "      <th>GSM1972910</th>\n",
       "      <th>GSM1972911</th>\n",
       "      <th>GSM1972912</th>\n",
       "      <th>GSM1972913</th>\n",
       "      <th>GSM1972914</th>\n",
       "      <th>GSM1972915</th>\n",
       "      <th>GSM1972916</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>3.527235</td>\n",
       "      <td>3.942945</td>\n",
       "      <td>3.671302</td>\n",
       "      <td>3.426890</td>\n",
       "      <td>4.123651</td>\n",
       "      <td>3.610940</td>\n",
       "      <td>3.854197</td>\n",
       "      <td>3.521270</td>\n",
       "      <td>3.702748</td>\n",
       "      <td>3.445015</td>\n",
       "      <td>...</td>\n",
       "      <td>3.452039</td>\n",
       "      <td>3.895396</td>\n",
       "      <td>3.793464</td>\n",
       "      <td>5.371881</td>\n",
       "      <td>3.303510</td>\n",
       "      <td>4.107289</td>\n",
       "      <td>3.424955</td>\n",
       "      <td>5.033872</td>\n",
       "      <td>3.835654</td>\n",
       "      <td>3.242641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>2.768377</td>\n",
       "      <td>3.032119</td>\n",
       "      <td>3.119842</td>\n",
       "      <td>3.150053</td>\n",
       "      <td>2.992697</td>\n",
       "      <td>3.102878</td>\n",
       "      <td>2.946598</td>\n",
       "      <td>2.842032</td>\n",
       "      <td>2.964397</td>\n",
       "      <td>3.112626</td>\n",
       "      <td>...</td>\n",
       "      <td>3.098948</td>\n",
       "      <td>2.943170</td>\n",
       "      <td>2.772752</td>\n",
       "      <td>2.973187</td>\n",
       "      <td>2.890580</td>\n",
       "      <td>3.198151</td>\n",
       "      <td>2.844137</td>\n",
       "      <td>2.723355</td>\n",
       "      <td>3.003175</td>\n",
       "      <td>3.194115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>6.851029</td>\n",
       "      <td>7.004668</td>\n",
       "      <td>6.664395</td>\n",
       "      <td>7.381467</td>\n",
       "      <td>7.472109</td>\n",
       "      <td>6.890454</td>\n",
       "      <td>6.771337</td>\n",
       "      <td>7.154243</td>\n",
       "      <td>7.811227</td>\n",
       "      <td>7.096253</td>\n",
       "      <td>...</td>\n",
       "      <td>6.461667</td>\n",
       "      <td>7.453001</td>\n",
       "      <td>6.572939</td>\n",
       "      <td>6.456971</td>\n",
       "      <td>6.982855</td>\n",
       "      <td>6.699946</td>\n",
       "      <td>6.735206</td>\n",
       "      <td>5.936276</td>\n",
       "      <td>6.416536</td>\n",
       "      <td>6.893309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>2.728755</td>\n",
       "      <td>2.687359</td>\n",
       "      <td>2.488260</td>\n",
       "      <td>2.902495</td>\n",
       "      <td>2.682164</td>\n",
       "      <td>2.781564</td>\n",
       "      <td>2.816926</td>\n",
       "      <td>2.812049</td>\n",
       "      <td>2.587465</td>\n",
       "      <td>2.584169</td>\n",
       "      <td>...</td>\n",
       "      <td>2.742156</td>\n",
       "      <td>4.398266</td>\n",
       "      <td>2.548673</td>\n",
       "      <td>2.714165</td>\n",
       "      <td>2.708875</td>\n",
       "      <td>2.685465</td>\n",
       "      <td>2.743197</td>\n",
       "      <td>2.585032</td>\n",
       "      <td>2.677635</td>\n",
       "      <td>2.906108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>5.240525</td>\n",
       "      <td>4.727202</td>\n",
       "      <td>5.289459</td>\n",
       "      <td>4.964199</td>\n",
       "      <td>4.803111</td>\n",
       "      <td>5.253464</td>\n",
       "      <td>5.141669</td>\n",
       "      <td>5.316193</td>\n",
       "      <td>5.079566</td>\n",
       "      <td>5.296088</td>\n",
       "      <td>...</td>\n",
       "      <td>4.850719</td>\n",
       "      <td>4.863967</td>\n",
       "      <td>5.088200</td>\n",
       "      <td>4.287189</td>\n",
       "      <td>4.269244</td>\n",
       "      <td>4.386475</td>\n",
       "      <td>3.986574</td>\n",
       "      <td>4.226469</td>\n",
       "      <td>5.957873</td>\n",
       "      <td>4.183791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>3.654807</td>\n",
       "      <td>3.037838</td>\n",
       "      <td>3.700867</td>\n",
       "      <td>3.236223</td>\n",
       "      <td>2.817592</td>\n",
       "      <td>3.508038</td>\n",
       "      <td>3.216719</td>\n",
       "      <td>3.302665</td>\n",
       "      <td>4.821776</td>\n",
       "      <td>3.345955</td>\n",
       "      <td>...</td>\n",
       "      <td>2.993568</td>\n",
       "      <td>3.124257</td>\n",
       "      <td>3.321288</td>\n",
       "      <td>3.786787</td>\n",
       "      <td>2.562922</td>\n",
       "      <td>2.999410</td>\n",
       "      <td>3.342488</td>\n",
       "      <td>4.395671</td>\n",
       "      <td>3.291762</td>\n",
       "      <td>3.143920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>8.106748</td>\n",
       "      <td>7.736487</td>\n",
       "      <td>7.929907</td>\n",
       "      <td>7.365184</td>\n",
       "      <td>7.334866</td>\n",
       "      <td>7.647479</td>\n",
       "      <td>8.404628</td>\n",
       "      <td>8.471322</td>\n",
       "      <td>7.707527</td>\n",
       "      <td>8.667955</td>\n",
       "      <td>...</td>\n",
       "      <td>8.148548</td>\n",
       "      <td>7.750400</td>\n",
       "      <td>7.696247</td>\n",
       "      <td>7.931895</td>\n",
       "      <td>7.958301</td>\n",
       "      <td>8.363429</td>\n",
       "      <td>8.426547</td>\n",
       "      <td>7.308236</td>\n",
       "      <td>8.403576</td>\n",
       "      <td>7.374028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>3.707414</td>\n",
       "      <td>3.824522</td>\n",
       "      <td>3.970674</td>\n",
       "      <td>3.953966</td>\n",
       "      <td>3.314021</td>\n",
       "      <td>3.750447</td>\n",
       "      <td>4.481623</td>\n",
       "      <td>5.350892</td>\n",
       "      <td>4.075008</td>\n",
       "      <td>4.400005</td>\n",
       "      <td>...</td>\n",
       "      <td>3.663827</td>\n",
       "      <td>3.892046</td>\n",
       "      <td>3.643871</td>\n",
       "      <td>3.635140</td>\n",
       "      <td>2.974060</td>\n",
       "      <td>3.826820</td>\n",
       "      <td>4.372133</td>\n",
       "      <td>3.554603</td>\n",
       "      <td>3.158875</td>\n",
       "      <td>3.565214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>4.953882</td>\n",
       "      <td>4.852440</td>\n",
       "      <td>4.961707</td>\n",
       "      <td>5.035332</td>\n",
       "      <td>5.086926</td>\n",
       "      <td>5.210269</td>\n",
       "      <td>5.084559</td>\n",
       "      <td>5.412141</td>\n",
       "      <td>4.688110</td>\n",
       "      <td>5.128178</td>\n",
       "      <td>...</td>\n",
       "      <td>5.296990</td>\n",
       "      <td>4.951781</td>\n",
       "      <td>4.795016</td>\n",
       "      <td>4.912349</td>\n",
       "      <td>5.306210</td>\n",
       "      <td>4.846485</td>\n",
       "      <td>5.563509</td>\n",
       "      <td>4.626624</td>\n",
       "      <td>5.128579</td>\n",
       "      <td>4.568179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>8.010302</td>\n",
       "      <td>7.458704</td>\n",
       "      <td>7.640166</td>\n",
       "      <td>7.455243</td>\n",
       "      <td>7.439838</td>\n",
       "      <td>8.055964</td>\n",
       "      <td>7.422100</td>\n",
       "      <td>7.491325</td>\n",
       "      <td>7.031681</td>\n",
       "      <td>7.838766</td>\n",
       "      <td>...</td>\n",
       "      <td>7.369171</td>\n",
       "      <td>7.530654</td>\n",
       "      <td>7.545323</td>\n",
       "      <td>7.729064</td>\n",
       "      <td>7.862185</td>\n",
       "      <td>7.626514</td>\n",
       "      <td>7.703143</td>\n",
       "      <td>7.479374</td>\n",
       "      <td>7.334361</td>\n",
       "      <td>7.149404</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15833 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM1972883  GSM1972884  GSM1972885  GSM1972886  GSM1972887  \\\n",
       "A1BG      3.527235    3.942945    3.671302    3.426890    4.123651   \n",
       "A1CF      2.768377    3.032119    3.119842    3.150053    2.992697   \n",
       "A2M       6.851029    7.004668    6.664395    7.381467    7.472109   \n",
       "A2ML1     2.728755    2.687359    2.488260    2.902495    2.682164   \n",
       "A4GALT    5.240525    4.727202    5.289459    4.964199    4.803111   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZYG11A    3.654807    3.037838    3.700867    3.236223    2.817592   \n",
       "ZYG11B    8.106748    7.736487    7.929907    7.365184    7.334866   \n",
       "ZYX       3.707414    3.824522    3.970674    3.953966    3.314021   \n",
       "ZZEF1     4.953882    4.852440    4.961707    5.035332    5.086926   \n",
       "ZZZ3      8.010302    7.458704    7.640166    7.455243    7.439838   \n",
       "\n",
       "        GSM1972888  GSM1972889  GSM1972890  GSM1972891  GSM1972892  ...  \\\n",
       "A1BG      3.610940    3.854197    3.521270    3.702748    3.445015  ...   \n",
       "A1CF      3.102878    2.946598    2.842032    2.964397    3.112626  ...   \n",
       "A2M       6.890454    6.771337    7.154243    7.811227    7.096253  ...   \n",
       "A2ML1     2.781564    2.816926    2.812049    2.587465    2.584169  ...   \n",
       "A4GALT    5.253464    5.141669    5.316193    5.079566    5.296088  ...   \n",
       "...            ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A    3.508038    3.216719    3.302665    4.821776    3.345955  ...   \n",
       "ZYG11B    7.647479    8.404628    8.471322    7.707527    8.667955  ...   \n",
       "ZYX       3.750447    4.481623    5.350892    4.075008    4.400005  ...   \n",
       "ZZEF1     5.210269    5.084559    5.412141    4.688110    5.128178  ...   \n",
       "ZZZ3      8.055964    7.422100    7.491325    7.031681    7.838766  ...   \n",
       "\n",
       "        GSM1972907  GSM1972908  GSM1972909  GSM1972910  GSM1972911  \\\n",
       "A1BG      3.452039    3.895396    3.793464    5.371881    3.303510   \n",
       "A1CF      3.098948    2.943170    2.772752    2.973187    2.890580   \n",
       "A2M       6.461667    7.453001    6.572939    6.456971    6.982855   \n",
       "A2ML1     2.742156    4.398266    2.548673    2.714165    2.708875   \n",
       "A4GALT    4.850719    4.863967    5.088200    4.287189    4.269244   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZYG11A    2.993568    3.124257    3.321288    3.786787    2.562922   \n",
       "ZYG11B    8.148548    7.750400    7.696247    7.931895    7.958301   \n",
       "ZYX       3.663827    3.892046    3.643871    3.635140    2.974060   \n",
       "ZZEF1     5.296990    4.951781    4.795016    4.912349    5.306210   \n",
       "ZZZ3      7.369171    7.530654    7.545323    7.729064    7.862185   \n",
       "\n",
       "        GSM1972912  GSM1972913  GSM1972914  GSM1972915  GSM1972916  \n",
       "A1BG      4.107289    3.424955    5.033872    3.835654    3.242641  \n",
       "A1CF      3.198151    2.844137    2.723355    3.003175    3.194115  \n",
       "A2M       6.699946    6.735206    5.936276    6.416536    6.893309  \n",
       "A2ML1     2.685465    2.743197    2.585032    2.677635    2.906108  \n",
       "A4GALT    4.386475    3.986574    4.226469    5.957873    4.183791  \n",
       "...            ...         ...         ...         ...         ...  \n",
       "ZYG11A    2.999410    3.342488    4.395671    3.291762    3.143920  \n",
       "ZYG11B    8.363429    8.426547    7.308236    8.403576    7.374028  \n",
       "ZYX       3.826820    4.372133    3.554603    3.158875    3.565214  \n",
       "ZZEF1     4.846485    5.563509    4.626624    5.128579    4.568179  \n",
       "ZZZ3      7.626514    7.703143    7.479374    7.334361    7.149404  \n",
       "\n",
       "[15833 rows x 34 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1972883</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.527235</td>\n",
       "      <td>2.768377</td>\n",
       "      <td>6.851029</td>\n",
       "      <td>2.728755</td>\n",
       "      <td>5.240525</td>\n",
       "      <td>3.477598</td>\n",
       "      <td>2.727000</td>\n",
       "      <td>5.676317</td>\n",
       "      <td>7.434579</td>\n",
       "      <td>...</td>\n",
       "      <td>8.509048</td>\n",
       "      <td>10.628571</td>\n",
       "      <td>5.649859</td>\n",
       "      <td>4.771590</td>\n",
       "      <td>6.386965</td>\n",
       "      <td>3.654807</td>\n",
       "      <td>8.106748</td>\n",
       "      <td>3.707414</td>\n",
       "      <td>4.953882</td>\n",
       "      <td>8.010302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972884</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.942945</td>\n",
       "      <td>3.032119</td>\n",
       "      <td>7.004668</td>\n",
       "      <td>2.687359</td>\n",
       "      <td>4.727202</td>\n",
       "      <td>3.252023</td>\n",
       "      <td>2.865537</td>\n",
       "      <td>5.516883</td>\n",
       "      <td>10.974107</td>\n",
       "      <td>...</td>\n",
       "      <td>7.971477</td>\n",
       "      <td>8.861770</td>\n",
       "      <td>5.179289</td>\n",
       "      <td>4.474295</td>\n",
       "      <td>5.412766</td>\n",
       "      <td>3.037838</td>\n",
       "      <td>7.736487</td>\n",
       "      <td>3.824522</td>\n",
       "      <td>4.852440</td>\n",
       "      <td>7.458704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972885</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.671302</td>\n",
       "      <td>3.119842</td>\n",
       "      <td>6.664395</td>\n",
       "      <td>2.488260</td>\n",
       "      <td>5.289459</td>\n",
       "      <td>3.501842</td>\n",
       "      <td>2.831561</td>\n",
       "      <td>5.064090</td>\n",
       "      <td>9.689137</td>\n",
       "      <td>...</td>\n",
       "      <td>7.075808</td>\n",
       "      <td>7.899708</td>\n",
       "      <td>5.244801</td>\n",
       "      <td>4.730381</td>\n",
       "      <td>5.308705</td>\n",
       "      <td>3.700867</td>\n",
       "      <td>7.929907</td>\n",
       "      <td>3.970674</td>\n",
       "      <td>4.961707</td>\n",
       "      <td>7.640166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972886</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.426890</td>\n",
       "      <td>3.150053</td>\n",
       "      <td>7.381467</td>\n",
       "      <td>2.902495</td>\n",
       "      <td>4.964199</td>\n",
       "      <td>3.465034</td>\n",
       "      <td>2.848093</td>\n",
       "      <td>5.627305</td>\n",
       "      <td>10.187094</td>\n",
       "      <td>...</td>\n",
       "      <td>8.322642</td>\n",
       "      <td>10.537770</td>\n",
       "      <td>5.080376</td>\n",
       "      <td>4.545587</td>\n",
       "      <td>6.046721</td>\n",
       "      <td>3.236223</td>\n",
       "      <td>7.365184</td>\n",
       "      <td>3.953966</td>\n",
       "      <td>5.035332</td>\n",
       "      <td>7.455243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972887</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.123651</td>\n",
       "      <td>2.992697</td>\n",
       "      <td>7.472109</td>\n",
       "      <td>2.682164</td>\n",
       "      <td>4.803111</td>\n",
       "      <td>3.165443</td>\n",
       "      <td>2.904853</td>\n",
       "      <td>5.582517</td>\n",
       "      <td>8.040422</td>\n",
       "      <td>...</td>\n",
       "      <td>8.334319</td>\n",
       "      <td>10.887228</td>\n",
       "      <td>5.542302</td>\n",
       "      <td>4.696713</td>\n",
       "      <td>6.357251</td>\n",
       "      <td>2.817592</td>\n",
       "      <td>7.334866</td>\n",
       "      <td>3.314021</td>\n",
       "      <td>5.086926</td>\n",
       "      <td>7.439838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972888</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.610940</td>\n",
       "      <td>3.102878</td>\n",
       "      <td>6.890454</td>\n",
       "      <td>2.781564</td>\n",
       "      <td>5.253464</td>\n",
       "      <td>3.304358</td>\n",
       "      <td>2.925750</td>\n",
       "      <td>5.896222</td>\n",
       "      <td>8.707148</td>\n",
       "      <td>...</td>\n",
       "      <td>7.457277</td>\n",
       "      <td>8.960566</td>\n",
       "      <td>4.444406</td>\n",
       "      <td>4.576363</td>\n",
       "      <td>5.658594</td>\n",
       "      <td>3.508038</td>\n",
       "      <td>7.647479</td>\n",
       "      <td>3.750447</td>\n",
       "      <td>5.210269</td>\n",
       "      <td>8.055964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972889</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.854197</td>\n",
       "      <td>2.946598</td>\n",
       "      <td>6.771337</td>\n",
       "      <td>2.816926</td>\n",
       "      <td>5.141669</td>\n",
       "      <td>3.375946</td>\n",
       "      <td>2.828268</td>\n",
       "      <td>5.466029</td>\n",
       "      <td>7.212664</td>\n",
       "      <td>...</td>\n",
       "      <td>6.303927</td>\n",
       "      <td>7.350314</td>\n",
       "      <td>6.185353</td>\n",
       "      <td>5.333719</td>\n",
       "      <td>6.151752</td>\n",
       "      <td>3.216719</td>\n",
       "      <td>8.404628</td>\n",
       "      <td>4.481623</td>\n",
       "      <td>5.084559</td>\n",
       "      <td>7.422100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972890</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.521270</td>\n",
       "      <td>2.842032</td>\n",
       "      <td>7.154243</td>\n",
       "      <td>2.812049</td>\n",
       "      <td>5.316193</td>\n",
       "      <td>3.417647</td>\n",
       "      <td>2.882909</td>\n",
       "      <td>5.840752</td>\n",
       "      <td>7.864725</td>\n",
       "      <td>...</td>\n",
       "      <td>6.783898</td>\n",
       "      <td>8.696888</td>\n",
       "      <td>5.681972</td>\n",
       "      <td>4.933937</td>\n",
       "      <td>5.540337</td>\n",
       "      <td>3.302665</td>\n",
       "      <td>8.471322</td>\n",
       "      <td>5.350892</td>\n",
       "      <td>5.412141</td>\n",
       "      <td>7.491325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972891</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.702748</td>\n",
       "      <td>2.964397</td>\n",
       "      <td>7.811227</td>\n",
       "      <td>2.587465</td>\n",
       "      <td>5.079566</td>\n",
       "      <td>3.310055</td>\n",
       "      <td>2.862623</td>\n",
       "      <td>5.371422</td>\n",
       "      <td>10.000973</td>\n",
       "      <td>...</td>\n",
       "      <td>7.685340</td>\n",
       "      <td>8.678479</td>\n",
       "      <td>5.005627</td>\n",
       "      <td>4.676198</td>\n",
       "      <td>5.377851</td>\n",
       "      <td>4.821776</td>\n",
       "      <td>7.707527</td>\n",
       "      <td>4.075008</td>\n",
       "      <td>4.688110</td>\n",
       "      <td>7.031681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972892</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.445015</td>\n",
       "      <td>3.112626</td>\n",
       "      <td>7.096253</td>\n",
       "      <td>2.584169</td>\n",
       "      <td>5.296088</td>\n",
       "      <td>3.364170</td>\n",
       "      <td>2.834300</td>\n",
       "      <td>5.265393</td>\n",
       "      <td>6.611711</td>\n",
       "      <td>...</td>\n",
       "      <td>6.465606</td>\n",
       "      <td>7.714065</td>\n",
       "      <td>5.411703</td>\n",
       "      <td>4.579231</td>\n",
       "      <td>6.845369</td>\n",
       "      <td>3.345955</td>\n",
       "      <td>8.667955</td>\n",
       "      <td>4.400005</td>\n",
       "      <td>5.128178</td>\n",
       "      <td>7.838766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972893</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.528796</td>\n",
       "      <td>3.003175</td>\n",
       "      <td>6.902870</td>\n",
       "      <td>2.741739</td>\n",
       "      <td>5.098354</td>\n",
       "      <td>3.291203</td>\n",
       "      <td>2.738746</td>\n",
       "      <td>5.274043</td>\n",
       "      <td>8.848443</td>\n",
       "      <td>...</td>\n",
       "      <td>7.359215</td>\n",
       "      <td>9.180896</td>\n",
       "      <td>4.847875</td>\n",
       "      <td>4.495096</td>\n",
       "      <td>5.462319</td>\n",
       "      <td>3.077862</td>\n",
       "      <td>8.482614</td>\n",
       "      <td>4.058140</td>\n",
       "      <td>7.392407</td>\n",
       "      <td>7.581649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972894</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.485716</td>\n",
       "      <td>2.927129</td>\n",
       "      <td>6.808653</td>\n",
       "      <td>2.465281</td>\n",
       "      <td>4.632869</td>\n",
       "      <td>3.284874</td>\n",
       "      <td>2.793344</td>\n",
       "      <td>5.414672</td>\n",
       "      <td>9.882016</td>\n",
       "      <td>...</td>\n",
       "      <td>8.500125</td>\n",
       "      <td>10.378777</td>\n",
       "      <td>4.668618</td>\n",
       "      <td>5.100604</td>\n",
       "      <td>6.085920</td>\n",
       "      <td>5.988908</td>\n",
       "      <td>8.574315</td>\n",
       "      <td>3.768274</td>\n",
       "      <td>5.912249</td>\n",
       "      <td>7.791607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972895</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.620551</td>\n",
       "      <td>2.940034</td>\n",
       "      <td>7.204748</td>\n",
       "      <td>2.579335</td>\n",
       "      <td>5.951053</td>\n",
       "      <td>3.191652</td>\n",
       "      <td>2.812921</td>\n",
       "      <td>5.958557</td>\n",
       "      <td>7.218691</td>\n",
       "      <td>...</td>\n",
       "      <td>8.014930</td>\n",
       "      <td>9.685383</td>\n",
       "      <td>4.482716</td>\n",
       "      <td>4.854579</td>\n",
       "      <td>5.526602</td>\n",
       "      <td>3.344583</td>\n",
       "      <td>7.020753</td>\n",
       "      <td>4.264689</td>\n",
       "      <td>5.136009</td>\n",
       "      <td>6.228846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972896</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.811414</td>\n",
       "      <td>2.982475</td>\n",
       "      <td>6.752328</td>\n",
       "      <td>3.074433</td>\n",
       "      <td>4.826542</td>\n",
       "      <td>2.948952</td>\n",
       "      <td>2.934928</td>\n",
       "      <td>5.606622</td>\n",
       "      <td>8.774746</td>\n",
       "      <td>...</td>\n",
       "      <td>7.391064</td>\n",
       "      <td>9.265674</td>\n",
       "      <td>5.102754</td>\n",
       "      <td>4.752586</td>\n",
       "      <td>5.317036</td>\n",
       "      <td>3.508281</td>\n",
       "      <td>7.994480</td>\n",
       "      <td>4.262824</td>\n",
       "      <td>4.629181</td>\n",
       "      <td>7.650382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972897</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.925597</td>\n",
       "      <td>3.022587</td>\n",
       "      <td>6.219221</td>\n",
       "      <td>2.511136</td>\n",
       "      <td>4.826542</td>\n",
       "      <td>3.745973</td>\n",
       "      <td>3.049097</td>\n",
       "      <td>5.510086</td>\n",
       "      <td>9.174977</td>\n",
       "      <td>...</td>\n",
       "      <td>7.543184</td>\n",
       "      <td>10.459155</td>\n",
       "      <td>5.124065</td>\n",
       "      <td>4.158999</td>\n",
       "      <td>6.260048</td>\n",
       "      <td>3.562693</td>\n",
       "      <td>7.824785</td>\n",
       "      <td>3.842163</td>\n",
       "      <td>5.531356</td>\n",
       "      <td>7.814737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972898</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.571770</td>\n",
       "      <td>2.801893</td>\n",
       "      <td>6.931969</td>\n",
       "      <td>2.671000</td>\n",
       "      <td>5.701167</td>\n",
       "      <td>3.114091</td>\n",
       "      <td>2.832078</td>\n",
       "      <td>5.485063</td>\n",
       "      <td>9.429489</td>\n",
       "      <td>...</td>\n",
       "      <td>7.182325</td>\n",
       "      <td>8.670419</td>\n",
       "      <td>5.542898</td>\n",
       "      <td>4.760132</td>\n",
       "      <td>5.678623</td>\n",
       "      <td>3.817049</td>\n",
       "      <td>8.485232</td>\n",
       "      <td>3.520020</td>\n",
       "      <td>5.467479</td>\n",
       "      <td>6.883867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972899</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.046221</td>\n",
       "      <td>2.775573</td>\n",
       "      <td>6.341712</td>\n",
       "      <td>2.630843</td>\n",
       "      <td>4.482458</td>\n",
       "      <td>3.379432</td>\n",
       "      <td>2.852100</td>\n",
       "      <td>5.498510</td>\n",
       "      <td>9.146923</td>\n",
       "      <td>...</td>\n",
       "      <td>9.056903</td>\n",
       "      <td>11.075373</td>\n",
       "      <td>4.974532</td>\n",
       "      <td>4.662365</td>\n",
       "      <td>5.890663</td>\n",
       "      <td>5.723219</td>\n",
       "      <td>9.098273</td>\n",
       "      <td>3.727714</td>\n",
       "      <td>6.052524</td>\n",
       "      <td>8.058514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972900</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.029233</td>\n",
       "      <td>2.978086</td>\n",
       "      <td>7.310797</td>\n",
       "      <td>2.648042</td>\n",
       "      <td>4.852013</td>\n",
       "      <td>3.386880</td>\n",
       "      <td>2.860698</td>\n",
       "      <td>5.360468</td>\n",
       "      <td>9.961633</td>\n",
       "      <td>...</td>\n",
       "      <td>8.191384</td>\n",
       "      <td>9.418835</td>\n",
       "      <td>5.136198</td>\n",
       "      <td>4.349113</td>\n",
       "      <td>5.534467</td>\n",
       "      <td>3.254438</td>\n",
       "      <td>7.878997</td>\n",
       "      <td>5.140663</td>\n",
       "      <td>4.873113</td>\n",
       "      <td>7.248824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972901</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.675049</td>\n",
       "      <td>2.869640</td>\n",
       "      <td>7.035170</td>\n",
       "      <td>2.519015</td>\n",
       "      <td>4.682002</td>\n",
       "      <td>3.072193</td>\n",
       "      <td>2.636666</td>\n",
       "      <td>5.528309</td>\n",
       "      <td>9.975805</td>\n",
       "      <td>...</td>\n",
       "      <td>7.755385</td>\n",
       "      <td>9.547054</td>\n",
       "      <td>5.121411</td>\n",
       "      <td>4.759683</td>\n",
       "      <td>5.584370</td>\n",
       "      <td>3.155685</td>\n",
       "      <td>8.103338</td>\n",
       "      <td>4.771235</td>\n",
       "      <td>4.860212</td>\n",
       "      <td>7.672669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972902</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.217550</td>\n",
       "      <td>3.006829</td>\n",
       "      <td>7.636877</td>\n",
       "      <td>2.878227</td>\n",
       "      <td>4.779781</td>\n",
       "      <td>3.312255</td>\n",
       "      <td>2.825914</td>\n",
       "      <td>5.358283</td>\n",
       "      <td>7.243100</td>\n",
       "      <td>...</td>\n",
       "      <td>6.634578</td>\n",
       "      <td>5.951790</td>\n",
       "      <td>5.484290</td>\n",
       "      <td>5.024246</td>\n",
       "      <td>5.166879</td>\n",
       "      <td>3.198343</td>\n",
       "      <td>7.725922</td>\n",
       "      <td>6.741842</td>\n",
       "      <td>4.921702</td>\n",
       "      <td>6.601325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972903</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.467937</td>\n",
       "      <td>2.801846</td>\n",
       "      <td>6.514929</td>\n",
       "      <td>2.705664</td>\n",
       "      <td>4.659055</td>\n",
       "      <td>3.404170</td>\n",
       "      <td>2.832140</td>\n",
       "      <td>5.687382</td>\n",
       "      <td>10.800238</td>\n",
       "      <td>...</td>\n",
       "      <td>6.833564</td>\n",
       "      <td>10.441410</td>\n",
       "      <td>5.017164</td>\n",
       "      <td>4.483262</td>\n",
       "      <td>5.974497</td>\n",
       "      <td>3.823357</td>\n",
       "      <td>7.556529</td>\n",
       "      <td>4.356279</td>\n",
       "      <td>5.868293</td>\n",
       "      <td>7.814908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972904</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.211319</td>\n",
       "      <td>3.306929</td>\n",
       "      <td>7.282108</td>\n",
       "      <td>2.805447</td>\n",
       "      <td>5.878285</td>\n",
       "      <td>2.884019</td>\n",
       "      <td>3.112322</td>\n",
       "      <td>5.565414</td>\n",
       "      <td>9.999651</td>\n",
       "      <td>...</td>\n",
       "      <td>6.214339</td>\n",
       "      <td>6.747455</td>\n",
       "      <td>4.025786</td>\n",
       "      <td>3.979136</td>\n",
       "      <td>5.750395</td>\n",
       "      <td>3.455691</td>\n",
       "      <td>6.746801</td>\n",
       "      <td>3.769678</td>\n",
       "      <td>5.159169</td>\n",
       "      <td>5.893547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972905</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.880413</td>\n",
       "      <td>2.842933</td>\n",
       "      <td>7.188670</td>\n",
       "      <td>2.632961</td>\n",
       "      <td>5.671016</td>\n",
       "      <td>3.254651</td>\n",
       "      <td>2.785316</td>\n",
       "      <td>5.062887</td>\n",
       "      <td>9.653624</td>\n",
       "      <td>...</td>\n",
       "      <td>7.109976</td>\n",
       "      <td>7.963831</td>\n",
       "      <td>4.723509</td>\n",
       "      <td>4.461222</td>\n",
       "      <td>5.951964</td>\n",
       "      <td>3.236735</td>\n",
       "      <td>7.422380</td>\n",
       "      <td>4.411756</td>\n",
       "      <td>4.976293</td>\n",
       "      <td>7.674682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972906</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.692146</td>\n",
       "      <td>2.992567</td>\n",
       "      <td>6.745187</td>\n",
       "      <td>2.685886</td>\n",
       "      <td>4.627328</td>\n",
       "      <td>3.202652</td>\n",
       "      <td>2.839997</td>\n",
       "      <td>5.455485</td>\n",
       "      <td>8.099521</td>\n",
       "      <td>...</td>\n",
       "      <td>7.408924</td>\n",
       "      <td>9.603512</td>\n",
       "      <td>5.521607</td>\n",
       "      <td>4.906920</td>\n",
       "      <td>5.767739</td>\n",
       "      <td>3.118649</td>\n",
       "      <td>8.200374</td>\n",
       "      <td>3.469593</td>\n",
       "      <td>4.964916</td>\n",
       "      <td>7.905258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972907</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.452039</td>\n",
       "      <td>3.098948</td>\n",
       "      <td>6.461667</td>\n",
       "      <td>2.742156</td>\n",
       "      <td>4.850719</td>\n",
       "      <td>3.214040</td>\n",
       "      <td>2.970536</td>\n",
       "      <td>5.656639</td>\n",
       "      <td>9.046744</td>\n",
       "      <td>...</td>\n",
       "      <td>6.675404</td>\n",
       "      <td>7.306308</td>\n",
       "      <td>5.481555</td>\n",
       "      <td>5.136076</td>\n",
       "      <td>6.272752</td>\n",
       "      <td>2.993568</td>\n",
       "      <td>8.148548</td>\n",
       "      <td>3.663827</td>\n",
       "      <td>5.296990</td>\n",
       "      <td>7.369171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972908</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.895396</td>\n",
       "      <td>2.943170</td>\n",
       "      <td>7.453001</td>\n",
       "      <td>4.398266</td>\n",
       "      <td>4.863967</td>\n",
       "      <td>3.359559</td>\n",
       "      <td>2.711770</td>\n",
       "      <td>5.877480</td>\n",
       "      <td>8.336773</td>\n",
       "      <td>...</td>\n",
       "      <td>6.757337</td>\n",
       "      <td>7.888227</td>\n",
       "      <td>6.173775</td>\n",
       "      <td>5.420720</td>\n",
       "      <td>6.340335</td>\n",
       "      <td>3.124257</td>\n",
       "      <td>7.750400</td>\n",
       "      <td>3.892046</td>\n",
       "      <td>4.951781</td>\n",
       "      <td>7.530654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972909</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.793464</td>\n",
       "      <td>2.772752</td>\n",
       "      <td>6.572939</td>\n",
       "      <td>2.548673</td>\n",
       "      <td>5.088200</td>\n",
       "      <td>3.487402</td>\n",
       "      <td>2.829673</td>\n",
       "      <td>4.764743</td>\n",
       "      <td>9.070172</td>\n",
       "      <td>...</td>\n",
       "      <td>7.824170</td>\n",
       "      <td>9.510114</td>\n",
       "      <td>4.283952</td>\n",
       "      <td>4.481998</td>\n",
       "      <td>5.219769</td>\n",
       "      <td>3.321288</td>\n",
       "      <td>7.696247</td>\n",
       "      <td>3.643871</td>\n",
       "      <td>4.795016</td>\n",
       "      <td>7.545323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972910</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.371881</td>\n",
       "      <td>2.973187</td>\n",
       "      <td>6.456971</td>\n",
       "      <td>2.714165</td>\n",
       "      <td>4.287189</td>\n",
       "      <td>3.268758</td>\n",
       "      <td>2.746400</td>\n",
       "      <td>5.270174</td>\n",
       "      <td>8.079602</td>\n",
       "      <td>...</td>\n",
       "      <td>8.712401</td>\n",
       "      <td>9.813376</td>\n",
       "      <td>5.093957</td>\n",
       "      <td>4.826237</td>\n",
       "      <td>5.730834</td>\n",
       "      <td>3.786787</td>\n",
       "      <td>7.931895</td>\n",
       "      <td>3.635140</td>\n",
       "      <td>4.912349</td>\n",
       "      <td>7.729064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972911</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.303510</td>\n",
       "      <td>2.890580</td>\n",
       "      <td>6.982855</td>\n",
       "      <td>2.708875</td>\n",
       "      <td>4.269244</td>\n",
       "      <td>3.286364</td>\n",
       "      <td>2.831909</td>\n",
       "      <td>5.172596</td>\n",
       "      <td>8.335202</td>\n",
       "      <td>...</td>\n",
       "      <td>8.503581</td>\n",
       "      <td>10.738182</td>\n",
       "      <td>4.552709</td>\n",
       "      <td>4.639886</td>\n",
       "      <td>6.308569</td>\n",
       "      <td>2.562922</td>\n",
       "      <td>7.958301</td>\n",
       "      <td>2.974060</td>\n",
       "      <td>5.306210</td>\n",
       "      <td>7.862185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972912</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4.107289</td>\n",
       "      <td>3.198151</td>\n",
       "      <td>6.699946</td>\n",
       "      <td>2.685465</td>\n",
       "      <td>4.386475</td>\n",
       "      <td>3.185804</td>\n",
       "      <td>3.225590</td>\n",
       "      <td>5.296312</td>\n",
       "      <td>10.383350</td>\n",
       "      <td>...</td>\n",
       "      <td>7.497950</td>\n",
       "      <td>10.051192</td>\n",
       "      <td>5.077114</td>\n",
       "      <td>4.686295</td>\n",
       "      <td>4.970305</td>\n",
       "      <td>2.999410</td>\n",
       "      <td>8.363429</td>\n",
       "      <td>3.826820</td>\n",
       "      <td>4.846485</td>\n",
       "      <td>7.626514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972913</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.424955</td>\n",
       "      <td>2.844137</td>\n",
       "      <td>6.735206</td>\n",
       "      <td>2.743197</td>\n",
       "      <td>3.986574</td>\n",
       "      <td>3.227203</td>\n",
       "      <td>2.905746</td>\n",
       "      <td>4.903879</td>\n",
       "      <td>6.997424</td>\n",
       "      <td>...</td>\n",
       "      <td>6.226428</td>\n",
       "      <td>7.643669</td>\n",
       "      <td>6.021632</td>\n",
       "      <td>5.073015</td>\n",
       "      <td>5.366580</td>\n",
       "      <td>3.342488</td>\n",
       "      <td>8.426547</td>\n",
       "      <td>4.372133</td>\n",
       "      <td>5.563509</td>\n",
       "      <td>7.703143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972914</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.033872</td>\n",
       "      <td>2.723355</td>\n",
       "      <td>5.936276</td>\n",
       "      <td>2.585032</td>\n",
       "      <td>4.226469</td>\n",
       "      <td>3.208216</td>\n",
       "      <td>2.929637</td>\n",
       "      <td>5.454201</td>\n",
       "      <td>7.444077</td>\n",
       "      <td>...</td>\n",
       "      <td>7.833585</td>\n",
       "      <td>10.973303</td>\n",
       "      <td>4.471790</td>\n",
       "      <td>4.174467</td>\n",
       "      <td>5.558383</td>\n",
       "      <td>4.395671</td>\n",
       "      <td>7.308236</td>\n",
       "      <td>3.554603</td>\n",
       "      <td>4.626624</td>\n",
       "      <td>7.479374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972915</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.835654</td>\n",
       "      <td>3.003175</td>\n",
       "      <td>6.416536</td>\n",
       "      <td>2.677635</td>\n",
       "      <td>5.957873</td>\n",
       "      <td>3.402576</td>\n",
       "      <td>2.760025</td>\n",
       "      <td>5.089679</td>\n",
       "      <td>8.152403</td>\n",
       "      <td>...</td>\n",
       "      <td>6.306715</td>\n",
       "      <td>7.551429</td>\n",
       "      <td>5.602928</td>\n",
       "      <td>5.085449</td>\n",
       "      <td>6.551380</td>\n",
       "      <td>3.291762</td>\n",
       "      <td>8.403576</td>\n",
       "      <td>3.158875</td>\n",
       "      <td>5.128579</td>\n",
       "      <td>7.334361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1972916</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.242641</td>\n",
       "      <td>3.194115</td>\n",
       "      <td>6.893309</td>\n",
       "      <td>2.906108</td>\n",
       "      <td>4.183791</td>\n",
       "      <td>3.438514</td>\n",
       "      <td>2.823032</td>\n",
       "      <td>4.941012</td>\n",
       "      <td>8.809485</td>\n",
       "      <td>...</td>\n",
       "      <td>8.373458</td>\n",
       "      <td>10.562473</td>\n",
       "      <td>4.774167</td>\n",
       "      <td>4.515114</td>\n",
       "      <td>5.721333</td>\n",
       "      <td>3.143920</td>\n",
       "      <td>7.374028</td>\n",
       "      <td>3.565214</td>\n",
       "      <td>4.568179</td>\n",
       "      <td>7.149404</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34 rows Ã— 15834 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Adrenocortical Cancer      A1BG      A1CF       A2M     A2ML1  \\\n",
       "GSM1972883                    1.0  3.527235  2.768377  6.851029  2.728755   \n",
       "GSM1972884                    1.0  3.942945  3.032119  7.004668  2.687359   \n",
       "GSM1972885                    1.0  3.671302  3.119842  6.664395  2.488260   \n",
       "GSM1972886                    1.0  3.426890  3.150053  7.381467  2.902495   \n",
       "GSM1972887                    1.0  4.123651  2.992697  7.472109  2.682164   \n",
       "GSM1972888                    1.0  3.610940  3.102878  6.890454  2.781564   \n",
       "GSM1972889                    1.0  3.854197  2.946598  6.771337  2.816926   \n",
       "GSM1972890                    1.0  3.521270  2.842032  7.154243  2.812049   \n",
       "GSM1972891                    1.0  3.702748  2.964397  7.811227  2.587465   \n",
       "GSM1972892                    1.0  3.445015  3.112626  7.096253  2.584169   \n",
       "GSM1972893                    1.0  3.528796  3.003175  6.902870  2.741739   \n",
       "GSM1972894                    1.0  3.485716  2.927129  6.808653  2.465281   \n",
       "GSM1972895                    1.0  3.620551  2.940034  7.204748  2.579335   \n",
       "GSM1972896                    1.0  3.811414  2.982475  6.752328  3.074433   \n",
       "GSM1972897                    1.0  4.925597  3.022587  6.219221  2.511136   \n",
       "GSM1972898                    1.0  5.571770  2.801893  6.931969  2.671000   \n",
       "GSM1972899                    1.0  4.046221  2.775573  6.341712  2.630843   \n",
       "GSM1972900                    1.0  4.029233  2.978086  7.310797  2.648042   \n",
       "GSM1972901                    1.0  3.675049  2.869640  7.035170  2.519015   \n",
       "GSM1972902                    1.0  4.217550  3.006829  7.636877  2.878227   \n",
       "GSM1972903                    1.0  3.467937  2.801846  6.514929  2.705664   \n",
       "GSM1972904                    1.0  4.211319  3.306929  7.282108  2.805447   \n",
       "GSM1972905                    1.0  3.880413  2.842933  7.188670  2.632961   \n",
       "GSM1972906                    1.0  3.692146  2.992567  6.745187  2.685886   \n",
       "GSM1972907                    1.0  3.452039  3.098948  6.461667  2.742156   \n",
       "GSM1972908                    1.0  3.895396  2.943170  7.453001  4.398266   \n",
       "GSM1972909                    1.0  3.793464  2.772752  6.572939  2.548673   \n",
       "GSM1972910                    1.0  5.371881  2.973187  6.456971  2.714165   \n",
       "GSM1972911                    1.0  3.303510  2.890580  6.982855  2.708875   \n",
       "GSM1972912                    1.0  4.107289  3.198151  6.699946  2.685465   \n",
       "GSM1972913                    1.0  3.424955  2.844137  6.735206  2.743197   \n",
       "GSM1972914                    1.0  5.033872  2.723355  5.936276  2.585032   \n",
       "GSM1972915                    1.0  3.835654  3.003175  6.416536  2.677635   \n",
       "GSM1972916                    1.0  3.242641  3.194115  6.893309  2.906108   \n",
       "\n",
       "              A4GALT     A4GNT      AAA1      AAAS       AACS  ...    ZWILCH  \\\n",
       "GSM1972883  5.240525  3.477598  2.727000  5.676317   7.434579  ...  8.509048   \n",
       "GSM1972884  4.727202  3.252023  2.865537  5.516883  10.974107  ...  7.971477   \n",
       "GSM1972885  5.289459  3.501842  2.831561  5.064090   9.689137  ...  7.075808   \n",
       "GSM1972886  4.964199  3.465034  2.848093  5.627305  10.187094  ...  8.322642   \n",
       "GSM1972887  4.803111  3.165443  2.904853  5.582517   8.040422  ...  8.334319   \n",
       "GSM1972888  5.253464  3.304358  2.925750  5.896222   8.707148  ...  7.457277   \n",
       "GSM1972889  5.141669  3.375946  2.828268  5.466029   7.212664  ...  6.303927   \n",
       "GSM1972890  5.316193  3.417647  2.882909  5.840752   7.864725  ...  6.783898   \n",
       "GSM1972891  5.079566  3.310055  2.862623  5.371422  10.000973  ...  7.685340   \n",
       "GSM1972892  5.296088  3.364170  2.834300  5.265393   6.611711  ...  6.465606   \n",
       "GSM1972893  5.098354  3.291203  2.738746  5.274043   8.848443  ...  7.359215   \n",
       "GSM1972894  4.632869  3.284874  2.793344  5.414672   9.882016  ...  8.500125   \n",
       "GSM1972895  5.951053  3.191652  2.812921  5.958557   7.218691  ...  8.014930   \n",
       "GSM1972896  4.826542  2.948952  2.934928  5.606622   8.774746  ...  7.391064   \n",
       "GSM1972897  4.826542  3.745973  3.049097  5.510086   9.174977  ...  7.543184   \n",
       "GSM1972898  5.701167  3.114091  2.832078  5.485063   9.429489  ...  7.182325   \n",
       "GSM1972899  4.482458  3.379432  2.852100  5.498510   9.146923  ...  9.056903   \n",
       "GSM1972900  4.852013  3.386880  2.860698  5.360468   9.961633  ...  8.191384   \n",
       "GSM1972901  4.682002  3.072193  2.636666  5.528309   9.975805  ...  7.755385   \n",
       "GSM1972902  4.779781  3.312255  2.825914  5.358283   7.243100  ...  6.634578   \n",
       "GSM1972903  4.659055  3.404170  2.832140  5.687382  10.800238  ...  6.833564   \n",
       "GSM1972904  5.878285  2.884019  3.112322  5.565414   9.999651  ...  6.214339   \n",
       "GSM1972905  5.671016  3.254651  2.785316  5.062887   9.653624  ...  7.109976   \n",
       "GSM1972906  4.627328  3.202652  2.839997  5.455485   8.099521  ...  7.408924   \n",
       "GSM1972907  4.850719  3.214040  2.970536  5.656639   9.046744  ...  6.675404   \n",
       "GSM1972908  4.863967  3.359559  2.711770  5.877480   8.336773  ...  6.757337   \n",
       "GSM1972909  5.088200  3.487402  2.829673  4.764743   9.070172  ...  7.824170   \n",
       "GSM1972910  4.287189  3.268758  2.746400  5.270174   8.079602  ...  8.712401   \n",
       "GSM1972911  4.269244  3.286364  2.831909  5.172596   8.335202  ...  8.503581   \n",
       "GSM1972912  4.386475  3.185804  3.225590  5.296312  10.383350  ...  7.497950   \n",
       "GSM1972913  3.986574  3.227203  2.905746  4.903879   6.997424  ...  6.226428   \n",
       "GSM1972914  4.226469  3.208216  2.929637  5.454201   7.444077  ...  7.833585   \n",
       "GSM1972915  5.957873  3.402576  2.760025  5.089679   8.152403  ...  6.306715   \n",
       "GSM1972916  4.183791  3.438514  2.823032  4.941012   8.809485  ...  8.373458   \n",
       "\n",
       "                ZWINT      ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B  \\\n",
       "GSM1972883  10.628571  5.649859  4.771590  6.386965  3.654807  8.106748   \n",
       "GSM1972884   8.861770  5.179289  4.474295  5.412766  3.037838  7.736487   \n",
       "GSM1972885   7.899708  5.244801  4.730381  5.308705  3.700867  7.929907   \n",
       "GSM1972886  10.537770  5.080376  4.545587  6.046721  3.236223  7.365184   \n",
       "GSM1972887  10.887228  5.542302  4.696713  6.357251  2.817592  7.334866   \n",
       "GSM1972888   8.960566  4.444406  4.576363  5.658594  3.508038  7.647479   \n",
       "GSM1972889   7.350314  6.185353  5.333719  6.151752  3.216719  8.404628   \n",
       "GSM1972890   8.696888  5.681972  4.933937  5.540337  3.302665  8.471322   \n",
       "GSM1972891   8.678479  5.005627  4.676198  5.377851  4.821776  7.707527   \n",
       "GSM1972892   7.714065  5.411703  4.579231  6.845369  3.345955  8.667955   \n",
       "GSM1972893   9.180896  4.847875  4.495096  5.462319  3.077862  8.482614   \n",
       "GSM1972894  10.378777  4.668618  5.100604  6.085920  5.988908  8.574315   \n",
       "GSM1972895   9.685383  4.482716  4.854579  5.526602  3.344583  7.020753   \n",
       "GSM1972896   9.265674  5.102754  4.752586  5.317036  3.508281  7.994480   \n",
       "GSM1972897  10.459155  5.124065  4.158999  6.260048  3.562693  7.824785   \n",
       "GSM1972898   8.670419  5.542898  4.760132  5.678623  3.817049  8.485232   \n",
       "GSM1972899  11.075373  4.974532  4.662365  5.890663  5.723219  9.098273   \n",
       "GSM1972900   9.418835  5.136198  4.349113  5.534467  3.254438  7.878997   \n",
       "GSM1972901   9.547054  5.121411  4.759683  5.584370  3.155685  8.103338   \n",
       "GSM1972902   5.951790  5.484290  5.024246  5.166879  3.198343  7.725922   \n",
       "GSM1972903  10.441410  5.017164  4.483262  5.974497  3.823357  7.556529   \n",
       "GSM1972904   6.747455  4.025786  3.979136  5.750395  3.455691  6.746801   \n",
       "GSM1972905   7.963831  4.723509  4.461222  5.951964  3.236735  7.422380   \n",
       "GSM1972906   9.603512  5.521607  4.906920  5.767739  3.118649  8.200374   \n",
       "GSM1972907   7.306308  5.481555  5.136076  6.272752  2.993568  8.148548   \n",
       "GSM1972908   7.888227  6.173775  5.420720  6.340335  3.124257  7.750400   \n",
       "GSM1972909   9.510114  4.283952  4.481998  5.219769  3.321288  7.696247   \n",
       "GSM1972910   9.813376  5.093957  4.826237  5.730834  3.786787  7.931895   \n",
       "GSM1972911  10.738182  4.552709  4.639886  6.308569  2.562922  7.958301   \n",
       "GSM1972912  10.051192  5.077114  4.686295  4.970305  2.999410  8.363429   \n",
       "GSM1972913   7.643669  6.021632  5.073015  5.366580  3.342488  8.426547   \n",
       "GSM1972914  10.973303  4.471790  4.174467  5.558383  4.395671  7.308236   \n",
       "GSM1972915   7.551429  5.602928  5.085449  6.551380  3.291762  8.403576   \n",
       "GSM1972916  10.562473  4.774167  4.515114  5.721333  3.143920  7.374028   \n",
       "\n",
       "                 ZYX     ZZEF1      ZZZ3  \n",
       "GSM1972883  3.707414  4.953882  8.010302  \n",
       "GSM1972884  3.824522  4.852440  7.458704  \n",
       "GSM1972885  3.970674  4.961707  7.640166  \n",
       "GSM1972886  3.953966  5.035332  7.455243  \n",
       "GSM1972887  3.314021  5.086926  7.439838  \n",
       "GSM1972888  3.750447  5.210269  8.055964  \n",
       "GSM1972889  4.481623  5.084559  7.422100  \n",
       "GSM1972890  5.350892  5.412141  7.491325  \n",
       "GSM1972891  4.075008  4.688110  7.031681  \n",
       "GSM1972892  4.400005  5.128178  7.838766  \n",
       "GSM1972893  4.058140  7.392407  7.581649  \n",
       "GSM1972894  3.768274  5.912249  7.791607  \n",
       "GSM1972895  4.264689  5.136009  6.228846  \n",
       "GSM1972896  4.262824  4.629181  7.650382  \n",
       "GSM1972897  3.842163  5.531356  7.814737  \n",
       "GSM1972898  3.520020  5.467479  6.883867  \n",
       "GSM1972899  3.727714  6.052524  8.058514  \n",
       "GSM1972900  5.140663  4.873113  7.248824  \n",
       "GSM1972901  4.771235  4.860212  7.672669  \n",
       "GSM1972902  6.741842  4.921702  6.601325  \n",
       "GSM1972903  4.356279  5.868293  7.814908  \n",
       "GSM1972904  3.769678  5.159169  5.893547  \n",
       "GSM1972905  4.411756  4.976293  7.674682  \n",
       "GSM1972906  3.469593  4.964916  7.905258  \n",
       "GSM1972907  3.663827  5.296990  7.369171  \n",
       "GSM1972908  3.892046  4.951781  7.530654  \n",
       "GSM1972909  3.643871  4.795016  7.545323  \n",
       "GSM1972910  3.635140  4.912349  7.729064  \n",
       "GSM1972911  2.974060  5.306210  7.862185  \n",
       "GSM1972912  3.826820  4.846485  7.626514  \n",
       "GSM1972913  4.372133  5.563509  7.703143  \n",
       "GSM1972914  3.554603  4.626624  7.479374  \n",
       "GSM1972915  3.158875  5.128579  7.334361  \n",
       "GSM1972916  3.565214  4.568179  7.149404  \n",
       "\n",
       "[34 rows x 15834 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 34 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 34 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"microRNA expression profile of pediatric adrenocortical tumors\"\n",
      "!Series_summary\t\"Here, we used a microarray technique to provide miRNA expression data of a set of 37 adrenocortical tumors (ACT) and 9 non-neoplastic adrenal controls from Brazilian patients assisted in two treatment centers in the state of SÃ£o Paulo (HC-FMRP-USP and Centro Infantil Boldrini of Campinas).\"\n",
      "!Series_overall_design\t\"We identified miRNA signatures associated with pediatric adrenocortical tumors and patients' outcome.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM5191580</th>\n",
       "      <th>GSM5191581</th>\n",
       "      <th>GSM5191582</th>\n",
       "      <th>GSM5191583</th>\n",
       "      <th>GSM5191584</th>\n",
       "      <th>GSM5191585</th>\n",
       "      <th>GSM5191586</th>\n",
       "      <th>GSM5191587</th>\n",
       "      <th>GSM5191588</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5191616</th>\n",
       "      <th>GSM5191617</th>\n",
       "      <th>GSM5191618</th>\n",
       "      <th>GSM5191619</th>\n",
       "      <th>GSM5191620</th>\n",
       "      <th>GSM5191621</th>\n",
       "      <th>GSM5191622</th>\n",
       "      <th>GSM5191623</th>\n",
       "      <th>GSM5191624</th>\n",
       "      <th>GSM5191625</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Tumor</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "      <td>tissue: Non-neoplastic Adrenal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age at diagnosis (months): 101</td>\n",
       "      <td>age at diagnosis (months): 13</td>\n",
       "      <td>age at diagnosis (months): 12</td>\n",
       "      <td>age at diagnosis (months): 29</td>\n",
       "      <td>age at diagnosis (months): 18</td>\n",
       "      <td>age at diagnosis (months): 137</td>\n",
       "      <td>age at diagnosis (months): 16</td>\n",
       "      <td>age at diagnosis (months): 95</td>\n",
       "      <td>age at diagnosis (months): 92</td>\n",
       "      <td>...</td>\n",
       "      <td>age at diagnosis (months): 34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sandrin stage: 2</td>\n",
       "      <td>sandrin stage: 1</td>\n",
       "      <td>sandrin stage: 2</td>\n",
       "      <td>sandrin stage: 1</td>\n",
       "      <td>sandrin stage: 1</td>\n",
       "      <td>sandrin stage: 2</td>\n",
       "      <td>sandrin stage: 1</td>\n",
       "      <td>sandrin stage: 4</td>\n",
       "      <td>sandrin stage: 4</td>\n",
       "      <td>...</td>\n",
       "      <td>sandrin stage: 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>metastasis: Present</td>\n",
       "      <td>metastasis: Present</td>\n",
       "      <td>...</td>\n",
       "      <td>metastasis: Absent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>relapse: present</td>\n",
       "      <td>relapse: absent</td>\n",
       "      <td>relapse: absent</td>\n",
       "      <td>relapse: present</td>\n",
       "      <td>relapse: absent</td>\n",
       "      <td>relapse: present</td>\n",
       "      <td>relapse: absent</td>\n",
       "      <td>relapse: present</td>\n",
       "      <td>relapse: present</td>\n",
       "      <td>...</td>\n",
       "      <td>relapse: absent</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>vital status: dead</td>\n",
       "      <td>vital status: alive</td>\n",
       "      <td>vital status: alive</td>\n",
       "      <td>vital status: dead</td>\n",
       "      <td>vital status: alive</td>\n",
       "      <td>vital status: alive</td>\n",
       "      <td>vital status: alive</td>\n",
       "      <td>vital status: dead</td>\n",
       "      <td>vital status: dead</td>\n",
       "      <td>...</td>\n",
       "      <td>vital status: alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows Ã— 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                      GSM5191580  \\\n",
       "0  !Sample_characteristics_ch1                   tissue: Tumor   \n",
       "1  !Sample_characteristics_ch1                  gender: Female   \n",
       "2  !Sample_characteristics_ch1  age at diagnosis (months): 101   \n",
       "3  !Sample_characteristics_ch1                sandrin stage: 2   \n",
       "4  !Sample_characteristics_ch1              metastasis: Absent   \n",
       "5  !Sample_characteristics_ch1                relapse: present   \n",
       "6  !Sample_characteristics_ch1              vital status: dead   \n",
       "\n",
       "                      GSM5191581                     GSM5191582  \\\n",
       "0                  tissue: Tumor                  tissue: Tumor   \n",
       "1                 gender: Female                   gender: Male   \n",
       "2  age at diagnosis (months): 13  age at diagnosis (months): 12   \n",
       "3               sandrin stage: 1               sandrin stage: 2   \n",
       "4             metastasis: Absent             metastasis: Absent   \n",
       "5                relapse: absent                relapse: absent   \n",
       "6            vital status: alive            vital status: alive   \n",
       "\n",
       "                      GSM5191583                     GSM5191584  \\\n",
       "0                  tissue: Tumor                  tissue: Tumor   \n",
       "1                 gender: Female                 gender: Female   \n",
       "2  age at diagnosis (months): 29  age at diagnosis (months): 18   \n",
       "3               sandrin stage: 1               sandrin stage: 1   \n",
       "4             metastasis: Absent             metastasis: Absent   \n",
       "5               relapse: present                relapse: absent   \n",
       "6             vital status: dead            vital status: alive   \n",
       "\n",
       "                       GSM5191585                     GSM5191586  \\\n",
       "0                   tissue: Tumor                  tissue: Tumor   \n",
       "1                  gender: Female                 gender: Female   \n",
       "2  age at diagnosis (months): 137  age at diagnosis (months): 16   \n",
       "3                sandrin stage: 2               sandrin stage: 1   \n",
       "4              metastasis: Absent             metastasis: Absent   \n",
       "5                relapse: present                relapse: absent   \n",
       "6             vital status: alive            vital status: alive   \n",
       "\n",
       "                      GSM5191587                     GSM5191588  ...  \\\n",
       "0                  tissue: Tumor                  tissue: Tumor  ...   \n",
       "1                 gender: Female                 gender: Female  ...   \n",
       "2  age at diagnosis (months): 95  age at diagnosis (months): 92  ...   \n",
       "3               sandrin stage: 4               sandrin stage: 4  ...   \n",
       "4            metastasis: Present            metastasis: Present  ...   \n",
       "5               relapse: present               relapse: present  ...   \n",
       "6             vital status: dead             vital status: dead  ...   \n",
       "\n",
       "                      GSM5191616                      GSM5191617  \\\n",
       "0                  tissue: Tumor  tissue: Non-neoplastic Adrenal   \n",
       "1                 gender: Female                             NaN   \n",
       "2  age at diagnosis (months): 34                             NaN   \n",
       "3               sandrin stage: 1                             NaN   \n",
       "4             metastasis: Absent                             NaN   \n",
       "5                relapse: absent                             NaN   \n",
       "6            vital status: alive                             NaN   \n",
       "\n",
       "                       GSM5191618                      GSM5191619  \\\n",
       "0  tissue: Non-neoplastic Adrenal  tissue: Non-neoplastic Adrenal   \n",
       "1                             NaN                             NaN   \n",
       "2                             NaN                             NaN   \n",
       "3                             NaN                             NaN   \n",
       "4                             NaN                             NaN   \n",
       "5                             NaN                             NaN   \n",
       "6                             NaN                             NaN   \n",
       "\n",
       "                       GSM5191620                      GSM5191621  \\\n",
       "0  tissue: Non-neoplastic Adrenal  tissue: Non-neoplastic Adrenal   \n",
       "1                             NaN                             NaN   \n",
       "2                             NaN                             NaN   \n",
       "3                             NaN                             NaN   \n",
       "4                             NaN                             NaN   \n",
       "5                             NaN                             NaN   \n",
       "6                             NaN                             NaN   \n",
       "\n",
       "                       GSM5191622                      GSM5191623  \\\n",
       "0  tissue: Non-neoplastic Adrenal  tissue: Non-neoplastic Adrenal   \n",
       "1                             NaN                             NaN   \n",
       "2                             NaN                             NaN   \n",
       "3                             NaN                             NaN   \n",
       "4                             NaN                             NaN   \n",
       "5                             NaN                             NaN   \n",
       "6                             NaN                             NaN   \n",
       "\n",
       "                       GSM5191624                      GSM5191625  \n",
       "0  tissue: Non-neoplastic Adrenal  tissue: Non-neoplastic Adrenal  \n",
       "1                             NaN                             NaN  \n",
       "2                             NaN                             NaN  \n",
       "3                             NaN                             NaN  \n",
       "4                             NaN                             NaN  \n",
       "5                             NaN                             NaN  \n",
       "6                             NaN                             NaN  \n",
       "\n",
       "[7 rows x 47 columns]"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No gene mapping\n",
    "cohort = accession_num = \"GSE169253\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'tissue: Tumor',\n",
       "       'tissue: Non-neoplastic Adrenal'], dtype=object)"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Tumor':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5191580</th>\n",
       "      <th>GSM5191581</th>\n",
       "      <th>GSM5191582</th>\n",
       "      <th>GSM5191583</th>\n",
       "      <th>GSM5191584</th>\n",
       "      <th>GSM5191585</th>\n",
       "      <th>GSM5191586</th>\n",
       "      <th>GSM5191587</th>\n",
       "      <th>GSM5191588</th>\n",
       "      <th>GSM5191589</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5191616</th>\n",
       "      <th>GSM5191617</th>\n",
       "      <th>GSM5191618</th>\n",
       "      <th>GSM5191619</th>\n",
       "      <th>GSM5191620</th>\n",
       "      <th>GSM5191621</th>\n",
       "      <th>GSM5191622</th>\n",
       "      <th>GSM5191623</th>\n",
       "      <th>GSM5191624</th>\n",
       "      <th>GSM5191625</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM5191580  GSM5191581  GSM5191582  GSM5191583  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM5191584  GSM5191585  GSM5191586  GSM5191587  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM5191588  GSM5191589  ...  GSM5191616  GSM5191617  \\\n",
       "Adrenocortical Cancer           0           0  ...           0           0   \n",
       "\n",
       "                       GSM5191618  GSM5191619  GSM5191620  GSM5191621  \\\n",
       "Adrenocortical Cancer           0           0           0           0   \n",
       "\n",
       "                       GSM5191622  GSM5191623  GSM5191624  GSM5191625  \n",
       "Adrenocortical Cancer           0           0           0           0  \n",
       "\n",
       "[1 rows x 46 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5191580</th>\n",
       "      <th>GSM5191581</th>\n",
       "      <th>GSM5191582</th>\n",
       "      <th>GSM5191583</th>\n",
       "      <th>GSM5191584</th>\n",
       "      <th>GSM5191585</th>\n",
       "      <th>GSM5191586</th>\n",
       "      <th>GSM5191587</th>\n",
       "      <th>GSM5191588</th>\n",
       "      <th>GSM5191589</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5191616</th>\n",
       "      <th>GSM5191617</th>\n",
       "      <th>GSM5191618</th>\n",
       "      <th>GSM5191619</th>\n",
       "      <th>GSM5191620</th>\n",
       "      <th>GSM5191621</th>\n",
       "      <th>GSM5191622</th>\n",
       "      <th>GSM5191623</th>\n",
       "      <th>GSM5191624</th>\n",
       "      <th>GSM5191625</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(-)3xSLv1</th>\n",
       "      <td>6.118841</td>\n",
       "      <td>6.157061</td>\n",
       "      <td>6.335244</td>\n",
       "      <td>6.169506</td>\n",
       "      <td>6.240245</td>\n",
       "      <td>6.182555</td>\n",
       "      <td>6.234796</td>\n",
       "      <td>6.238807</td>\n",
       "      <td>6.326222</td>\n",
       "      <td>6.116632</td>\n",
       "      <td>...</td>\n",
       "      <td>6.275570</td>\n",
       "      <td>6.359939</td>\n",
       "      <td>6.111164</td>\n",
       "      <td>6.431596</td>\n",
       "      <td>6.424909</td>\n",
       "      <td>6.592276</td>\n",
       "      <td>6.277234</td>\n",
       "      <td>6.378845</td>\n",
       "      <td>6.257708</td>\n",
       "      <td>6.248218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_25_P00010019</th>\n",
       "      <td>6.299119</td>\n",
       "      <td>8.455914</td>\n",
       "      <td>10.030732</td>\n",
       "      <td>8.614498</td>\n",
       "      <td>8.260569</td>\n",
       "      <td>6.156289</td>\n",
       "      <td>9.278782</td>\n",
       "      <td>9.181840</td>\n",
       "      <td>9.148011</td>\n",
       "      <td>7.061471</td>\n",
       "      <td>...</td>\n",
       "      <td>8.409829</td>\n",
       "      <td>6.443060</td>\n",
       "      <td>7.243682</td>\n",
       "      <td>6.375311</td>\n",
       "      <td>6.409273</td>\n",
       "      <td>6.444193</td>\n",
       "      <td>6.223879</td>\n",
       "      <td>6.458491</td>\n",
       "      <td>6.147493</td>\n",
       "      <td>6.358181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_25_P00010020</th>\n",
       "      <td>6.083133</td>\n",
       "      <td>7.270414</td>\n",
       "      <td>8.654528</td>\n",
       "      <td>7.275610</td>\n",
       "      <td>7.228507</td>\n",
       "      <td>6.077727</td>\n",
       "      <td>7.992475</td>\n",
       "      <td>8.030643</td>\n",
       "      <td>7.993448</td>\n",
       "      <td>6.422413</td>\n",
       "      <td>...</td>\n",
       "      <td>7.339864</td>\n",
       "      <td>6.325392</td>\n",
       "      <td>6.493788</td>\n",
       "      <td>6.214819</td>\n",
       "      <td>6.177738</td>\n",
       "      <td>6.231950</td>\n",
       "      <td>6.111671</td>\n",
       "      <td>6.213652</td>\n",
       "      <td>6.293622</td>\n",
       "      <td>6.265370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_25_P00010021</th>\n",
       "      <td>6.073802</td>\n",
       "      <td>6.217121</td>\n",
       "      <td>7.033983</td>\n",
       "      <td>6.302130</td>\n",
       "      <td>6.375919</td>\n",
       "      <td>6.176956</td>\n",
       "      <td>6.654326</td>\n",
       "      <td>7.128662</td>\n",
       "      <td>6.347221</td>\n",
       "      <td>6.175253</td>\n",
       "      <td>...</td>\n",
       "      <td>6.516075</td>\n",
       "      <td>6.196105</td>\n",
       "      <td>6.116335</td>\n",
       "      <td>6.210674</td>\n",
       "      <td>6.313971</td>\n",
       "      <td>6.436155</td>\n",
       "      <td>6.091414</td>\n",
       "      <td>6.143457</td>\n",
       "      <td>6.371254</td>\n",
       "      <td>6.194190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_25_P00010023</th>\n",
       "      <td>6.255995</td>\n",
       "      <td>8.369547</td>\n",
       "      <td>8.928362</td>\n",
       "      <td>6.998827</td>\n",
       "      <td>7.619062</td>\n",
       "      <td>6.196523</td>\n",
       "      <td>7.914588</td>\n",
       "      <td>9.366658</td>\n",
       "      <td>8.620604</td>\n",
       "      <td>6.521392</td>\n",
       "      <td>...</td>\n",
       "      <td>7.135663</td>\n",
       "      <td>6.169613</td>\n",
       "      <td>6.351142</td>\n",
       "      <td>6.261724</td>\n",
       "      <td>6.388602</td>\n",
       "      <td>6.527635</td>\n",
       "      <td>6.128198</td>\n",
       "      <td>6.246212</td>\n",
       "      <td>6.260787</td>\n",
       "      <td>6.352774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC1_00000215</th>\n",
       "      <td>6.267363</td>\n",
       "      <td>6.148166</td>\n",
       "      <td>6.403182</td>\n",
       "      <td>6.150068</td>\n",
       "      <td>6.271648</td>\n",
       "      <td>6.231808</td>\n",
       "      <td>6.283637</td>\n",
       "      <td>6.285456</td>\n",
       "      <td>6.302437</td>\n",
       "      <td>6.176041</td>\n",
       "      <td>...</td>\n",
       "      <td>6.357121</td>\n",
       "      <td>6.406852</td>\n",
       "      <td>6.180986</td>\n",
       "      <td>6.733853</td>\n",
       "      <td>6.699904</td>\n",
       "      <td>6.815947</td>\n",
       "      <td>6.316242</td>\n",
       "      <td>6.685653</td>\n",
       "      <td>6.484752</td>\n",
       "      <td>6.380198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00079215</th>\n",
       "      <td>6.097034</td>\n",
       "      <td>6.094926</td>\n",
       "      <td>6.254911</td>\n",
       "      <td>6.095361</td>\n",
       "      <td>6.160075</td>\n",
       "      <td>6.150899</td>\n",
       "      <td>6.162629</td>\n",
       "      <td>6.161875</td>\n",
       "      <td>6.243301</td>\n",
       "      <td>6.088003</td>\n",
       "      <td>...</td>\n",
       "      <td>6.222913</td>\n",
       "      <td>6.285008</td>\n",
       "      <td>6.098394</td>\n",
       "      <td>6.350706</td>\n",
       "      <td>6.190702</td>\n",
       "      <td>6.370142</td>\n",
       "      <td>6.253163</td>\n",
       "      <td>6.251502</td>\n",
       "      <td>6.238752</td>\n",
       "      <td>6.184274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00092197</th>\n",
       "      <td>6.228357</td>\n",
       "      <td>6.112642</td>\n",
       "      <td>6.317175</td>\n",
       "      <td>6.146962</td>\n",
       "      <td>6.252238</td>\n",
       "      <td>6.217719</td>\n",
       "      <td>6.215052</td>\n",
       "      <td>6.227773</td>\n",
       "      <td>6.254591</td>\n",
       "      <td>6.104834</td>\n",
       "      <td>...</td>\n",
       "      <td>6.312603</td>\n",
       "      <td>6.341783</td>\n",
       "      <td>6.105114</td>\n",
       "      <td>6.381271</td>\n",
       "      <td>6.328268</td>\n",
       "      <td>6.633423</td>\n",
       "      <td>6.198914</td>\n",
       "      <td>6.304024</td>\n",
       "      <td>6.285849</td>\n",
       "      <td>6.366904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00106057</th>\n",
       "      <td>6.137844</td>\n",
       "      <td>6.139085</td>\n",
       "      <td>6.449337</td>\n",
       "      <td>6.161739</td>\n",
       "      <td>6.262088</td>\n",
       "      <td>6.271165</td>\n",
       "      <td>6.260400</td>\n",
       "      <td>6.256848</td>\n",
       "      <td>6.328337</td>\n",
       "      <td>6.124522</td>\n",
       "      <td>...</td>\n",
       "      <td>6.368682</td>\n",
       "      <td>6.474196</td>\n",
       "      <td>6.168280</td>\n",
       "      <td>6.876222</td>\n",
       "      <td>6.709560</td>\n",
       "      <td>7.003939</td>\n",
       "      <td>6.514869</td>\n",
       "      <td>6.711632</td>\n",
       "      <td>6.346856</td>\n",
       "      <td>6.351947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC2_00122731</th>\n",
       "      <td>6.142575</td>\n",
       "      <td>6.099026</td>\n",
       "      <td>6.308705</td>\n",
       "      <td>6.105705</td>\n",
       "      <td>6.227261</td>\n",
       "      <td>6.154260</td>\n",
       "      <td>6.215354</td>\n",
       "      <td>6.249252</td>\n",
       "      <td>6.258837</td>\n",
       "      <td>6.124059</td>\n",
       "      <td>...</td>\n",
       "      <td>6.239852</td>\n",
       "      <td>6.281127</td>\n",
       "      <td>6.168355</td>\n",
       "      <td>6.423401</td>\n",
       "      <td>6.362061</td>\n",
       "      <td>6.578652</td>\n",
       "      <td>6.226190</td>\n",
       "      <td>6.461765</td>\n",
       "      <td>6.273917</td>\n",
       "      <td>6.251256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2696 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GSM5191580  GSM5191581  GSM5191582  GSM5191583  GSM5191584  \\\n",
       "ID                                                                           \n",
       "(-)3xSLv1         6.118841    6.157061    6.335244    6.169506    6.240245   \n",
       "A_25_P00010019    6.299119    8.455914   10.030732    8.614498    8.260569   \n",
       "A_25_P00010020    6.083133    7.270414    8.654528    7.275610    7.228507   \n",
       "A_25_P00010021    6.073802    6.217121    7.033983    6.302130    6.375919   \n",
       "A_25_P00010023    6.255995    8.369547    8.928362    6.998827    7.619062   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "NC1_00000215      6.267363    6.148166    6.403182    6.150068    6.271648   \n",
       "NC2_00079215      6.097034    6.094926    6.254911    6.095361    6.160075   \n",
       "NC2_00092197      6.228357    6.112642    6.317175    6.146962    6.252238   \n",
       "NC2_00106057      6.137844    6.139085    6.449337    6.161739    6.262088   \n",
       "NC2_00122731      6.142575    6.099026    6.308705    6.105705    6.227261   \n",
       "\n",
       "                GSM5191585  GSM5191586  GSM5191587  GSM5191588  GSM5191589  \\\n",
       "ID                                                                           \n",
       "(-)3xSLv1         6.182555    6.234796    6.238807    6.326222    6.116632   \n",
       "A_25_P00010019    6.156289    9.278782    9.181840    9.148011    7.061471   \n",
       "A_25_P00010020    6.077727    7.992475    8.030643    7.993448    6.422413   \n",
       "A_25_P00010021    6.176956    6.654326    7.128662    6.347221    6.175253   \n",
       "A_25_P00010023    6.196523    7.914588    9.366658    8.620604    6.521392   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "NC1_00000215      6.231808    6.283637    6.285456    6.302437    6.176041   \n",
       "NC2_00079215      6.150899    6.162629    6.161875    6.243301    6.088003   \n",
       "NC2_00092197      6.217719    6.215052    6.227773    6.254591    6.104834   \n",
       "NC2_00106057      6.271165    6.260400    6.256848    6.328337    6.124522   \n",
       "NC2_00122731      6.154260    6.215354    6.249252    6.258837    6.124059   \n",
       "\n",
       "                ...  GSM5191616  GSM5191617  GSM5191618  GSM5191619  \\\n",
       "ID              ...                                                   \n",
       "(-)3xSLv1       ...    6.275570    6.359939    6.111164    6.431596   \n",
       "A_25_P00010019  ...    8.409829    6.443060    7.243682    6.375311   \n",
       "A_25_P00010020  ...    7.339864    6.325392    6.493788    6.214819   \n",
       "A_25_P00010021  ...    6.516075    6.196105    6.116335    6.210674   \n",
       "A_25_P00010023  ...    7.135663    6.169613    6.351142    6.261724   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "NC1_00000215    ...    6.357121    6.406852    6.180986    6.733853   \n",
       "NC2_00079215    ...    6.222913    6.285008    6.098394    6.350706   \n",
       "NC2_00092197    ...    6.312603    6.341783    6.105114    6.381271   \n",
       "NC2_00106057    ...    6.368682    6.474196    6.168280    6.876222   \n",
       "NC2_00122731    ...    6.239852    6.281127    6.168355    6.423401   \n",
       "\n",
       "                GSM5191620  GSM5191621  GSM5191622  GSM5191623  GSM5191624  \\\n",
       "ID                                                                           \n",
       "(-)3xSLv1         6.424909    6.592276    6.277234    6.378845    6.257708   \n",
       "A_25_P00010019    6.409273    6.444193    6.223879    6.458491    6.147493   \n",
       "A_25_P00010020    6.177738    6.231950    6.111671    6.213652    6.293622   \n",
       "A_25_P00010021    6.313971    6.436155    6.091414    6.143457    6.371254   \n",
       "A_25_P00010023    6.388602    6.527635    6.128198    6.246212    6.260787   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "NC1_00000215      6.699904    6.815947    6.316242    6.685653    6.484752   \n",
       "NC2_00079215      6.190702    6.370142    6.253163    6.251502    6.238752   \n",
       "NC2_00092197      6.328268    6.633423    6.198914    6.304024    6.285849   \n",
       "NC2_00106057      6.709560    7.003939    6.514869    6.711632    6.346856   \n",
       "NC2_00122731      6.362061    6.578652    6.226190    6.461765    6.273917   \n",
       "\n",
       "                GSM5191625  \n",
       "ID                          \n",
       "(-)3xSLv1         6.248218  \n",
       "A_25_P00010019    6.358181  \n",
       "A_25_P00010020    6.265370  \n",
       "A_25_P00010021    6.194190  \n",
       "A_25_P00010023    6.352774  \n",
       "...                    ...  \n",
       "NC1_00000215      6.380198  \n",
       "NC2_00079215      6.184274  \n",
       "NC2_00092197      6.366904  \n",
       "NC2_00106057      6.351947  \n",
       "NC2_00122731      6.251256  \n",
       "\n",
       "[2696 rows x 46 columns]"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['(-)3xSLv1', 'A_25_P00010019', 'A_25_P00010020', 'A_25_P00010021', 'A_25_P00010023'], 'miRNA_ID': [nan, 'hsa-miR-329', 'hsa-miR-329', 'hsa-miR-655', 'hsa-miR-369-3p'], 'Accession_String': [nan, 'mir|hsa-miR-329|mir|MIMAT0001629', 'mir|hsa-miR-329|mir|MIMAT0001629', 'mir|hsa-miR-655|mir|MIMAT0003331', 'mir|hsa-miR-369-3p|mir|MIMAT0000721'], 'SPOT_ID': ['NegativeControl', nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['(-)3xSLv1', 'A_25_P00010019', 'A_25_P00010020', 'A_25_P00010021', 'A_25_P00010023', 'A_25_P00010037', 'A_25_P00010038', 'A_25_P00010039', 'A_25_P00010040', 'A_25_P00010041', 'A_25_P00010042', 'A_25_P00010043', 'A_25_P00010044', 'A_25_P00010047', 'A_25_P00010048', 'A_25_P00010053', 'A_25_P00010054', 'A_25_P00010062', 'A_25_P00010063', 'A_25_P00010070']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['(-)3xSLv1', 'A_25_P00010019', 'A_25_P00010020', 'A_25_P00010021', 'A_25_P00010023'], 'miRNA_ID': [nan, 'hsa-miR-329', 'hsa-miR-329', 'hsa-miR-655', 'hsa-miR-369-3p'], 'Accession_String': [nan, 'mir|hsa-miR-329|mir|MIMAT0001629', 'mir|hsa-miR-329|mir|MIMAT0001629', 'mir|hsa-miR-655|mir|MIMAT0003331', 'mir|hsa-miR-369-3p|mir|MIMAT0000721'], 'SPOT_ID': ['NegativeControl', nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>Accession_String</th>\n",
       "      <th>SPOT_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(-)3xSLv1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NegativeControl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_25_P00010019</td>\n",
       "      <td>hsa-miR-329</td>\n",
       "      <td>mir|hsa-miR-329|mir|MIMAT0001629</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_25_P00010020</td>\n",
       "      <td>hsa-miR-329</td>\n",
       "      <td>mir|hsa-miR-329|mir|MIMAT0001629</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_25_P00010021</td>\n",
       "      <td>hsa-miR-655</td>\n",
       "      <td>mir|hsa-miR-655|mir|MIMAT0003331</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_25_P00010023</td>\n",
       "      <td>hsa-miR-369-3p</td>\n",
       "      <td>mir|hsa-miR-369-3p|mir|MIMAT0000721</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126792</th>\n",
       "      <td>A_25_P00012251</td>\n",
       "      <td>6.060039071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126793</th>\n",
       "      <td>A_25_P00013186</td>\n",
       "      <td>6.008080699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126794</th>\n",
       "      <td>A_25_P00013312</td>\n",
       "      <td>5.980089392</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126795</th>\n",
       "      <td>A_25_P00012716</td>\n",
       "      <td>6.063907312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126796</th>\n",
       "      <td>A_25_P00013797</td>\n",
       "      <td>5.953854308</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126797 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID        miRNA_ID                     Accession_String  \\\n",
       "0            (-)3xSLv1             NaN                                  NaN   \n",
       "1       A_25_P00010019     hsa-miR-329     mir|hsa-miR-329|mir|MIMAT0001629   \n",
       "2       A_25_P00010020     hsa-miR-329     mir|hsa-miR-329|mir|MIMAT0001629   \n",
       "3       A_25_P00010021     hsa-miR-655     mir|hsa-miR-655|mir|MIMAT0003331   \n",
       "4       A_25_P00010023  hsa-miR-369-3p  mir|hsa-miR-369-3p|mir|MIMAT0000721   \n",
       "...                ...             ...                                  ...   \n",
       "126792  A_25_P00012251     6.060039071                                  NaN   \n",
       "126793  A_25_P00013186     6.008080699                                  NaN   \n",
       "126794  A_25_P00013312     5.980089392                                  NaN   \n",
       "126795  A_25_P00012716     6.063907312                                  NaN   \n",
       "126796  A_25_P00013797     5.953854308                                  NaN   \n",
       "\n",
       "                SPOT_ID  \n",
       "0       NegativeControl  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  \n",
       "...                 ...  \n",
       "126792              NaN  \n",
       "126793              NaN  \n",
       "126794              NaN  \n",
       "126795              NaN  \n",
       "126796              NaN  \n",
       "\n",
       "[126797 rows x 4 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'miRNA_ID', 'Accession_String', 'SPOT_ID'], dtype='object')"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Cancer Cells Hijack PRC2 to Modify Multiple Cytokine Pathways\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1652385</th>\n",
       "      <th>GSM1652386</th>\n",
       "      <th>GSM1652387</th>\n",
       "      <th>GSM1652388</th>\n",
       "      <th>GSM1652389</th>\n",
       "      <th>GSM1652390</th>\n",
       "      <th>GSM1652391</th>\n",
       "      <th>GSM1652392</th>\n",
       "      <th>GSM1652393</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1652399</th>\n",
       "      <th>GSM1652400</th>\n",
       "      <th>GSM1652401</th>\n",
       "      <th>GSM1652402</th>\n",
       "      <th>GSM1652403</th>\n",
       "      <th>GSM1652404</th>\n",
       "      <th>GSM1652405</th>\n",
       "      <th>GSM1652406</th>\n",
       "      <th>GSM1652407</th>\n",
       "      <th>GSM1652408</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>...</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "      <td>cell line: SW-13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession        GSM1652385        GSM1652386  \\\n",
       "0  !Sample_characteristics_ch1  cell line: SW-13  cell line: SW-13   \n",
       "\n",
       "         GSM1652387        GSM1652388        GSM1652389        GSM1652390  \\\n",
       "0  cell line: SW-13  cell line: SW-13  cell line: SW-13  cell line: SW-13   \n",
       "\n",
       "         GSM1652391        GSM1652392        GSM1652393  ...  \\\n",
       "0  cell line: SW-13  cell line: SW-13  cell line: SW-13  ...   \n",
       "\n",
       "         GSM1652399        GSM1652400        GSM1652401        GSM1652402  \\\n",
       "0  cell line: SW-13  cell line: SW-13  cell line: SW-13  cell line: SW-13   \n",
       "\n",
       "         GSM1652403        GSM1652404        GSM1652405        GSM1652406  \\\n",
       "0  cell line: SW-13  cell line: SW-13  cell line: SW-13  cell line: SW-13   \n",
       "\n",
       "         GSM1652407        GSM1652408  \n",
       "0  cell line: SW-13  cell line: SW-13  \n",
       "\n",
       "[1 rows x 25 columns]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No trait \n",
    "cohort = accession_num = \"GSE67766\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "ename": "BadGzipFile",
     "evalue": "Not a gzipped file (b'!S')",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mBadGzipFile\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[223], line 12\u001b[0m\n\u001b[0;32m      9\u001b[0m background_prefixes \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Series_title\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Series_summary\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Series_overall_design\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m     10\u001b[0m clinical_prefixes \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Sample_geo_accession\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Sample_characteristics_ch1\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m---> 12\u001b[0m background_info, clinical_data \u001b[38;5;241m=\u001b[39m \u001b[43mget_background_and_clinical_data\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmatrix_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbackground_prefixes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mclinical_prefixes\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     13\u001b[0m \u001b[38;5;28mprint\u001b[39m(background_info)\n\u001b[0;32m     15\u001b[0m clinical_data\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:120\u001b[0m, in \u001b[0;36mget_background_and_clinical_data\u001b[1;34m(file_path, prefixes_a, prefixes_b)\u001b[0m\n\u001b[0;32m    116\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_background_and_clinical_data\u001b[39m(file_path,\n\u001b[0;32m    117\u001b[0m                                      prefixes_a\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Series_title\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Series_summary\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Series_overall_design\u001b[39m\u001b[38;5;124m'\u001b[39m],\n\u001b[0;32m    118\u001b[0m                                      prefixes_b\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Sample_geo_accession\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Sample_characteristics_ch1\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    119\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\u001b[39;00m\n\u001b[1;32m--> 120\u001b[0m     background_info, clinical_data \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_content_by_prefix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_a\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_b\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munselect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    121\u001b[0m \u001b[43m                                                              \u001b[49m\u001b[43msource_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfile\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    122\u001b[0m \u001b[43m                                                              \u001b[49m\u001b[43mreturn_df_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreturn_df_b\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    123\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m background_info, clinical_data\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:95\u001b[0m, in \u001b[0;36mfilter_content_by_prefix\u001b[1;34m(source, prefixes_a, prefixes_b, unselect, source_type, return_df_a, return_df_b)\u001b[0m\n\u001b[0;32m     92\u001b[0m     prefix_set_b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m(prefixes_b)\n\u001b[0;32m     94\u001b[0m \u001b[38;5;66;03m# Use generator to get lines\u001b[39;00m\n\u001b[1;32m---> 95\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m line_generator(source, source_type):\n\u001b[0;32m     96\u001b[0m     matched_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;28many\u001b[39m(line\u001b[38;5;241m.\u001b[39mstartswith(prefix) \u001b[38;5;28;01mfor\u001b[39;00m prefix \u001b[38;5;129;01min\u001b[39;00m prefix_set_a)\n\u001b[0;32m     97\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m matched_a \u001b[38;5;241m!=\u001b[39m unselect:\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:55\u001b[0m, in \u001b[0;36mline_generator\u001b[1;34m(source, source_type)\u001b[0m\n\u001b[0;32m     53\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m source_type \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfile\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m     54\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m gzip\u001b[38;5;241m.\u001b[39mopen(source, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrt\u001b[39m\u001b[38;5;124m'\u001b[39m, encoding\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mutf-8\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[1;32m---> 55\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m f:\n\u001b[0;32m     56\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m line\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[0;32m     57\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m source_type \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstring\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "File \u001b[1;32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\\Lib\\gzip.py:314\u001b[0m, in \u001b[0;36mGzipFile.read1\u001b[1;34m(self, size)\u001b[0m\n\u001b[0;32m    312\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m size \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m    313\u001b[0m     size \u001b[38;5;241m=\u001b[39m io\u001b[38;5;241m.\u001b[39mDEFAULT_BUFFER_SIZE\n\u001b[1;32m--> 314\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer\u001b[38;5;241m.\u001b[39mread1(size)\n",
      "File \u001b[1;32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\\Lib\\_compression.py:68\u001b[0m, in \u001b[0;36mDecompressReader.readinto\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m     66\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mreadinto\u001b[39m(\u001b[38;5;28mself\u001b[39m, b):\n\u001b[0;32m     67\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mmemoryview\u001b[39m(b) \u001b[38;5;28;01mas\u001b[39;00m view, view\u001b[38;5;241m.\u001b[39mcast(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mB\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m byte_view:\n\u001b[1;32m---> 68\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mbyte_view\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     69\u001b[0m         byte_view[:\u001b[38;5;28mlen\u001b[39m(data)] \u001b[38;5;241m=\u001b[39m data\n\u001b[0;32m     70\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(data)\n",
      "File \u001b[1;32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\\Lib\\gzip.py:499\u001b[0m, in \u001b[0;36m_GzipReader.read\u001b[1;34m(self, size)\u001b[0m\n\u001b[0;32m    495\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_new_member:\n\u001b[0;32m    496\u001b[0m     \u001b[38;5;66;03m# If the _new_member flag is set, we have to\u001b[39;00m\n\u001b[0;32m    497\u001b[0m     \u001b[38;5;66;03m# jump to the next member, if there is one.\u001b[39;00m\n\u001b[0;32m    498\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_init_read()\n\u001b[1;32m--> 499\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_gzip_header\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[0;32m    500\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pos\n\u001b[0;32m    501\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[1;32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\\Lib\\gzip.py:468\u001b[0m, in \u001b[0;36m_GzipReader._read_gzip_header\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    467\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_gzip_header\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m--> 468\u001b[0m     last_mtime \u001b[38;5;241m=\u001b[39m \u001b[43m_read_gzip_header\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    469\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m last_mtime \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    470\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[1;32mC:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.11_3.11.2288.0_x64__qbz5n2kfra8p0\\Lib\\gzip.py:428\u001b[0m, in \u001b[0;36m_read_gzip_header\u001b[1;34m(fp)\u001b[0m\n\u001b[0;32m    425\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    427\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m magic \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\037\u001b[39;00m\u001b[38;5;130;01m\\213\u001b[39;00m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m--> 428\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m BadGzipFile(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNot a gzipped file (\u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m%\u001b[39m magic)\n\u001b[0;32m    430\u001b[0m (method, flag, last_mtime) \u001b[38;5;241m=\u001b[39m struct\u001b[38;5;241m.\u001b[39munpack(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m<BBIxx\u001b[39m\u001b[38;5;124m\"\u001b[39m, _read_exact(fp, \u001b[38;5;241m8\u001b[39m))\n\u001b[0;32m    431\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m8\u001b[39m:\n",
      "\u001b[1;31mBadGzipFile\u001b[0m: Not a gzipped file (b'!S')"
     ]
    }
   ],
   "source": [
    "# age + sex (problematic)\n",
    "cohort = accession_num = \"GSE68606\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Advancing a Clinically Relevant Perspective of the Clonal Nature of Cancer\"\n",
      "!Series_summary\t\"We used DNA content-based flow cytometry to distinguish and isolate nuclei from clonal populations in primary tissues from three disparate cancers with variable clinical histories. We then developed a methodology to adapt flow cytometrically purified nuclei samples for use with whole genome technologies including aCGH and next generation sequencing (NGS). Our results demonstrate that selected aberrations in the genomes of distinct clonal populations in each patient create clinically relevant contexts at least with respect to the cancer types profiled in this study.\"\n",
      "!Series_overall_design\t\"We applied DNA content based flow sorting to isolate the nuclei of clonal populations from tumor biopsies. Genomic DNA from each sorted population was amplified with phi29 polymerase. A 1ug aliquot of each amplified sample was digested with DNAse 1 then labeled with Cy5 using a Klenow-based commercial kit (Invitrogen). Each sample was hybridized with a pooled normal (46,XX) reference (Promega) to Agilent 244k CGH arrays. The use of highly purified objectively defined flow sorted populations provides high definition genomic profiles of clonal populations from pancreatic adenocarcinomas (PA), adrenal cortical carcinomas (ACC), and prostate adenocarcinomas (PC).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                 GSM540550  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540551  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540552  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540553  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540554  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540555  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540556  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540557  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540558  ...  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma  ...   \n",
       "\n",
       "                            GSM540601                           GSM540602  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Adrenal Cortical Carcinoma   \n",
       "\n",
       "                            GSM540603                   GSM540604  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540605                   GSM540606  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540607                   GSM540608  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540609                   GSM540610  \n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE21660\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'tissue: Pancreatic Ductal Adenocarcinoma',\n",
       "       'tissue: Adrenal Cortical Carcinoma', 'tissue: Prostate Carcinoma'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Adrenal Cortical Carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "\n",
       "                       GSM540555  GSM540556  GSM540557  GSM540558  GSM540559  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "\n",
       "                       ...  GSM540601  GSM540602  GSM540603  GSM540604  \\\n",
       "Adrenocortical Cancer  ...          1          1          1          0   \n",
       "\n",
       "                       GSM540605  GSM540606  GSM540607  GSM540608  GSM540609  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "\n",
       "                       GSM540610  \n",
       "Adrenocortical Cancer          0  \n",
       "\n",
       "[1 rows x 61 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.00599</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>-0.0779</td>\n",
       "      <td>0.01020</td>\n",
       "      <td>-0.00558</td>\n",
       "      <td>-0.0871</td>\n",
       "      <td>-0.0895</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.00695</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0563</td>\n",
       "      <td>0.002070</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.00841</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.01280</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>-0.000229</td>\n",
       "      <td>-0.0148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1730</td>\n",
       "      <td>-0.43800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.6140</td>\n",
       "      <td>-0.0190</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.5670</td>\n",
       "      <td>-0.004620</td>\n",
       "      <td>-0.4700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.2000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3190</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.4420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.6850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.19900</td>\n",
       "      <td>0.1570</td>\n",
       "      <td>0.2230</td>\n",
       "      <td>0.01230</td>\n",
       "      <td>0.13200</td>\n",
       "      <td>0.0682</td>\n",
       "      <td>0.1330</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>0.03560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1920</td>\n",
       "      <td>0.3650</td>\n",
       "      <td>0.364000</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>-0.24800</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>-0.08650</td>\n",
       "      <td>0.0336</td>\n",
       "      <td>0.167000</td>\n",
       "      <td>0.0337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.1050</td>\n",
       "      <td>-0.06010</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.1780</td>\n",
       "      <td>-0.00527</td>\n",
       "      <td>0.00652</td>\n",
       "      <td>-0.1610</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>-0.1520</td>\n",
       "      <td>-0.02830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0991</td>\n",
       "      <td>-0.0738</td>\n",
       "      <td>-0.058400</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>-0.06350</td>\n",
       "      <td>-0.1470</td>\n",
       "      <td>-0.24700</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>-0.115000</td>\n",
       "      <td>0.0135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243500</th>\n",
       "      <td>-0.2810</td>\n",
       "      <td>-0.34800</td>\n",
       "      <td>-0.3390</td>\n",
       "      <td>-0.2320</td>\n",
       "      <td>0.02470</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>-0.2440</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>-0.1790</td>\n",
       "      <td>-0.34600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>-0.1520</td>\n",
       "      <td>-0.211000</td>\n",
       "      <td>-0.0559</td>\n",
       "      <td>0.08870</td>\n",
       "      <td>-0.0608</td>\n",
       "      <td>0.00423</td>\n",
       "      <td>-0.4000</td>\n",
       "      <td>-0.069500</td>\n",
       "      <td>-0.0305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243501</th>\n",
       "      <td>-0.1770</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.2670</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.28800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0433</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2160</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.122000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243502</th>\n",
       "      <td>-0.5300</td>\n",
       "      <td>1.30000</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.66700</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.2540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.1480</td>\n",
       "      <td>-0.302000</td>\n",
       "      <td>-0.9700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243503</th>\n",
       "      <td>0.0344</td>\n",
       "      <td>0.02240</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>-0.0692</td>\n",
       "      <td>0.03500</td>\n",
       "      <td>0.01120</td>\n",
       "      <td>-0.0938</td>\n",
       "      <td>-0.0330</td>\n",
       "      <td>-0.1030</td>\n",
       "      <td>0.00613</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0709</td>\n",
       "      <td>0.0626</td>\n",
       "      <td>-0.000328</td>\n",
       "      <td>0.0182</td>\n",
       "      <td>0.01140</td>\n",
       "      <td>0.0342</td>\n",
       "      <td>-0.03030</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>-0.0138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243504</th>\n",
       "      <td>0.0434</td>\n",
       "      <td>0.01580</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>0.05200</td>\n",
       "      <td>0.00951</td>\n",
       "      <td>-0.0797</td>\n",
       "      <td>-0.0113</td>\n",
       "      <td>-0.1080</td>\n",
       "      <td>-0.00240</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0639</td>\n",
       "      <td>0.0483</td>\n",
       "      <td>-0.002440</td>\n",
       "      <td>0.0539</td>\n",
       "      <td>0.02960</td>\n",
       "      <td>0.0264</td>\n",
       "      <td>-0.00543</td>\n",
       "      <td>0.0700</td>\n",
       "      <td>-0.012400</td>\n",
       "      <td>-0.0203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243429 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "ID                                                                         \n",
       "1          0.0201    0.00599     0.0117    -0.0779    0.01020   -0.00558   \n",
       "2          0.0000    0.00000     0.0000    -0.1730   -0.43800    0.00000   \n",
       "3         -0.2000    0.00000     0.0000     0.0000    0.00000    0.00000   \n",
       "4          0.1210    0.19900     0.1570     0.2230    0.01230    0.13200   \n",
       "5         -0.1050   -0.06010    -0.1140    -0.1780   -0.00527    0.00652   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "243500    -0.2810   -0.34800    -0.3390    -0.2320    0.02470    0.02690   \n",
       "243501    -0.1770    0.00000    -0.2670     0.0000   -0.28800    0.00000   \n",
       "243502    -0.5300    1.30000     0.0179     0.0000   -0.66700    0.00000   \n",
       "243503     0.0344    0.02240     0.0275    -0.0692    0.03500    0.01120   \n",
       "243504     0.0434    0.01580     0.0122    -0.0646    0.05200    0.00951   \n",
       "\n",
       "        GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  GSM540602  \\\n",
       "ID                                                  ...                         \n",
       "1         -0.0871    -0.0895    -0.1140   -0.00695  ...     0.0592     0.0563   \n",
       "2          0.0000    -0.6140    -0.0190    0.00000  ...     0.0000     0.0000   \n",
       "3          0.0000    -0.3190     0.0000    0.00000  ...     0.0000     0.0000   \n",
       "4          0.0682     0.1330     0.1480    0.03560  ...     0.1920     0.3650   \n",
       "5         -0.1610    -0.1200    -0.1520   -0.02830  ...     0.0991    -0.0738   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "243500    -0.2440     0.0160    -0.1790   -0.34600  ...     0.1480    -0.1520   \n",
       "243501     0.0000     0.0000     0.0433    0.00000  ...     0.2160     0.0000   \n",
       "243502     0.0000     0.0000     0.0000    0.00000  ...     0.0000    -0.2540   \n",
       "243503    -0.0938    -0.0330    -0.1030    0.00613  ...     0.0709     0.0626   \n",
       "243504    -0.0797    -0.0113    -0.1080   -0.00240  ...     0.0639     0.0483   \n",
       "\n",
       "        GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  GSM540608  \\\n",
       "ID                                                                         \n",
       "1        0.002070     0.0208    0.00841     0.0122   -0.01280     0.0748   \n",
       "2        0.000000     0.0000    0.00000     0.0000    0.00000    -0.5670   \n",
       "3        0.000000     0.0000    0.00000     0.0000    0.00000    -0.4420   \n",
       "4        0.364000     0.0321   -0.24800     0.0240   -0.08650     0.0336   \n",
       "5       -0.058400     0.0120   -0.06350    -0.1470   -0.24700     0.1170   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "243500  -0.211000    -0.0559    0.08870    -0.0608    0.00423    -0.4000   \n",
       "243501   0.122000     0.0000    0.00000     0.0000    0.00000     0.0000   \n",
       "243502   0.000000     0.0000    0.00000     0.0000    0.00000    -0.1480   \n",
       "243503  -0.000328     0.0182    0.01140     0.0342   -0.03030     0.0543   \n",
       "243504  -0.002440     0.0539    0.02960     0.0264   -0.00543     0.0700   \n",
       "\n",
       "        GSM540609  GSM540610  \n",
       "ID                            \n",
       "1       -0.000229    -0.0148  \n",
       "2       -0.004620    -0.4700  \n",
       "3        0.000000    -0.6850  \n",
       "4        0.167000     0.0337  \n",
       "5       -0.115000     0.0135  \n",
       "...           ...        ...  \n",
       "243500  -0.069500    -0.0305  \n",
       "243501   0.000000    -1.0400  \n",
       "243502  -0.302000    -0.9700  \n",
       "243503   0.014000    -0.0138  \n",
       "243504  -0.012400    -0.0203  \n",
       "\n",
       "[243429 rows x 61 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1', '2', '3', '4', '5'], 'COL': ['267', '267', '267', '267', '267'], 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0], 'SPOT_ID': ['HsCGHBrightCorner', 'DarkCorner', 'DarkCorner', 'A_16_P20527812', 'A_16_P01708709'], 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'], 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'], 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'], 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'], 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'], 'CHROMOSOMAL_LOCATION': [nan, nan, nan, 'chr16:076331867-076331926', 'chr7:047626734-047626793'], 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'], 'DESCRIPTION': [nan, nan, nan, nan, 'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'], 'GB_RANGE': [nan, nan, nan, 'NC_000016.8[076331867..076331926]', 'NC_000007.11[047626734..047626793]']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '13', '14', '15', '16', '17', '18', '19', '20', '21']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1', '2', '3', '4', '5'], 'COL': ['267', '267', '267', '267', '267'], 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0], 'SPOT_ID': ['HsCGHBrightCorner', 'DarkCorner', 'DarkCorner', 'A_16_P20527812', 'A_16_P01708709'], 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'], 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'], 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'], 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'], 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'], 'CHROMOSOMAL_LOCATION': [nan, nan, nan, 'chr16:076331867-076331926', 'chr7:047626734-047626793'], 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'], 'DESCRIPTION': [nan, nan, nan, nan, 'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'], 'GB_RANGE': [nan, nan, nan, 'NC_000016.8[076331867..076331926]', 'NC_000007.11[047626734..047626793]']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>COL</th>\n",
       "      <th>ROW</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>CONTROL_TYPE</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ACCESSION_STRING</th>\n",
       "      <th>CHROMOSOMAL_LOCATION</th>\n",
       "      <th>CYTOBAND</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>GB_RANGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>267</td>\n",
       "      <td>912.0</td>\n",
       "      <td>HsCGHBrightCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>267</td>\n",
       "      <td>910.0</td>\n",
       "      <td>DarkCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>267</td>\n",
       "      <td>908.0</td>\n",
       "      <td>DarkCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>267</td>\n",
       "      <td>906.0</td>\n",
       "      <td>A_16_P20527812</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16:076331867-076331926</td>\n",
       "      <td>hs|q23.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000016.8[076331867..076331926]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>267</td>\n",
       "      <td>904.0</td>\n",
       "      <td>A_16_P01708709</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NM_138295</td>\n",
       "      <td>PKD1L1</td>\n",
       "      <td>polycystic kidney disease 1 like 1</td>\n",
       "      <td>ref|NM_138295|ref|NM_025031</td>\n",
       "      <td>chr7:047626734-047626793</td>\n",
       "      <td>hs|p12.3</td>\n",
       "      <td>Homo sapiens polycystic kidney disease 1 like ...</td>\n",
       "      <td>NC_000007.11[047626734..047626793]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092790</th>\n",
       "      <td>243500</td>\n",
       "      <td>-3.05E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092791</th>\n",
       "      <td>243501</td>\n",
       "      <td>-1.04E+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092792</th>\n",
       "      <td>243502</td>\n",
       "      <td>-9.70E-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092793</th>\n",
       "      <td>243503</td>\n",
       "      <td>-1.38E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092794</th>\n",
       "      <td>243504</td>\n",
       "      <td>-2.03E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15092795 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID        COL    ROW            SPOT_ID CONTROL_TYPE     GB_ACC  \\\n",
       "0              1        267  912.0  HsCGHBrightCorner          pos        NaN   \n",
       "1              2        267  910.0         DarkCorner          pos        NaN   \n",
       "2              3        267  908.0         DarkCorner          pos        NaN   \n",
       "3              4        267  906.0     A_16_P20527812        FALSE        NaN   \n",
       "4              5        267  904.0     A_16_P01708709        FALSE  NM_138295   \n",
       "...          ...        ...    ...                ...          ...        ...   \n",
       "15092790  243500  -3.05E-02    NaN                NaN          NaN        NaN   \n",
       "15092791  243501  -1.04E+00    NaN                NaN          NaN        NaN   \n",
       "15092792  243502  -9.70E-01    NaN                NaN          NaN        NaN   \n",
       "15092793  243503  -1.38E-02    NaN                NaN          NaN        NaN   \n",
       "15092794  243504  -2.03E-02    NaN                NaN          NaN        NaN   \n",
       "\n",
       "         GENE_SYMBOL                           GENE_NAME  \\\n",
       "0                NaN                                 NaN   \n",
       "1                NaN                                 NaN   \n",
       "2                NaN                                 NaN   \n",
       "3                NaN                                 NaN   \n",
       "4             PKD1L1  polycystic kidney disease 1 like 1   \n",
       "...              ...                                 ...   \n",
       "15092790         NaN                                 NaN   \n",
       "15092791         NaN                                 NaN   \n",
       "15092792         NaN                                 NaN   \n",
       "15092793         NaN                                 NaN   \n",
       "15092794         NaN                                 NaN   \n",
       "\n",
       "                     ACCESSION_STRING       CHROMOSOMAL_LOCATION  CYTOBAND  \\\n",
       "0                                 NaN                        NaN       NaN   \n",
       "1                                 NaN                        NaN       NaN   \n",
       "2                                 NaN                        NaN       NaN   \n",
       "3                                 NaN  chr16:076331867-076331926  hs|q23.1   \n",
       "4         ref|NM_138295|ref|NM_025031   chr7:047626734-047626793  hs|p12.3   \n",
       "...                               ...                        ...       ...   \n",
       "15092790                          NaN                        NaN       NaN   \n",
       "15092791                          NaN                        NaN       NaN   \n",
       "15092792                          NaN                        NaN       NaN   \n",
       "15092793                          NaN                        NaN       NaN   \n",
       "15092794                          NaN                        NaN       NaN   \n",
       "\n",
       "                                                DESCRIPTION  \\\n",
       "0                                                       NaN   \n",
       "1                                                       NaN   \n",
       "2                                                       NaN   \n",
       "3                                                       NaN   \n",
       "4         Homo sapiens polycystic kidney disease 1 like ...   \n",
       "...                                                     ...   \n",
       "15092790                                                NaN   \n",
       "15092791                                                NaN   \n",
       "15092792                                                NaN   \n",
       "15092793                                                NaN   \n",
       "15092794                                                NaN   \n",
       "\n",
       "                                    GB_RANGE  \n",
       "0                                        NaN  \n",
       "1                                        NaN  \n",
       "2                                        NaN  \n",
       "3          NC_000016.8[076331867..076331926]  \n",
       "4         NC_000007.11[047626734..047626793]  \n",
       "...                                      ...  \n",
       "15092790                                 NaN  \n",
       "15092791                                 NaN  \n",
       "15092792                                 NaN  \n",
       "15092793                                 NaN  \n",
       "15092794                                 NaN  \n",
       "\n",
       "[15092795 rows x 13 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'COL', 'ROW', 'SPOT_ID', 'CONTROL_TYPE', 'GB_ACC', 'GENE_SYMBOL',\n",
       "       'GENE_NAME', 'ACCESSION_STRING', 'CHROMOSOMAL_LOCATION', 'CYTOBAND',\n",
       "       'DESCRIPTION', 'GB_RANGE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GENE_SYMBOL'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>-0.091366</td>\n",
       "      <td>-0.009954</td>\n",
       "      <td>0.064026</td>\n",
       "      <td>0.223229</td>\n",
       "      <td>0.015514</td>\n",
       "      <td>0.133943</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.036996</td>\n",
       "      <td>-0.109300</td>\n",
       "      <td>0.058823</td>\n",
       "      <td>...</td>\n",
       "      <td>0.152000</td>\n",
       "      <td>0.040586</td>\n",
       "      <td>0.114029</td>\n",
       "      <td>-0.021514</td>\n",
       "      <td>0.110771</td>\n",
       "      <td>0.135557</td>\n",
       "      <td>0.125829</td>\n",
       "      <td>0.009514</td>\n",
       "      <td>0.174671</td>\n",
       "      <td>0.049243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15E1.2</th>\n",
       "      <td>-0.018767</td>\n",
       "      <td>-0.036233</td>\n",
       "      <td>0.126533</td>\n",
       "      <td>0.015967</td>\n",
       "      <td>-0.001400</td>\n",
       "      <td>-0.229000</td>\n",
       "      <td>0.095173</td>\n",
       "      <td>-0.137293</td>\n",
       "      <td>-0.145300</td>\n",
       "      <td>0.137067</td>\n",
       "      <td>...</td>\n",
       "      <td>0.146700</td>\n",
       "      <td>0.267000</td>\n",
       "      <td>0.207333</td>\n",
       "      <td>0.039400</td>\n",
       "      <td>-0.047812</td>\n",
       "      <td>0.011360</td>\n",
       "      <td>-0.109767</td>\n",
       "      <td>0.300333</td>\n",
       "      <td>0.013167</td>\n",
       "      <td>-0.008633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2'-PDE</th>\n",
       "      <td>-0.095650</td>\n",
       "      <td>-0.038200</td>\n",
       "      <td>-0.273500</td>\n",
       "      <td>-0.188500</td>\n",
       "      <td>0.077200</td>\n",
       "      <td>-0.170500</td>\n",
       "      <td>-0.023250</td>\n",
       "      <td>-0.265000</td>\n",
       "      <td>-0.224050</td>\n",
       "      <td>0.028700</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.189500</td>\n",
       "      <td>-0.117450</td>\n",
       "      <td>-0.166800</td>\n",
       "      <td>-0.037500</td>\n",
       "      <td>-0.123500</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>-0.111050</td>\n",
       "      <td>0.182000</td>\n",
       "      <td>-0.032200</td>\n",
       "      <td>-0.052400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76P</th>\n",
       "      <td>-0.177225</td>\n",
       "      <td>-0.016975</td>\n",
       "      <td>-0.015500</td>\n",
       "      <td>-0.230250</td>\n",
       "      <td>-0.381500</td>\n",
       "      <td>-0.132625</td>\n",
       "      <td>-0.074625</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>-0.158050</td>\n",
       "      <td>-0.093265</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067575</td>\n",
       "      <td>0.068450</td>\n",
       "      <td>-0.156500</td>\n",
       "      <td>0.048800</td>\n",
       "      <td>-0.054375</td>\n",
       "      <td>-0.010340</td>\n",
       "      <td>-0.110975</td>\n",
       "      <td>0.131200</td>\n",
       "      <td>-0.107950</td>\n",
       "      <td>-0.035025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7A5</th>\n",
       "      <td>-0.060111</td>\n",
       "      <td>0.072104</td>\n",
       "      <td>0.055131</td>\n",
       "      <td>-0.020300</td>\n",
       "      <td>0.201556</td>\n",
       "      <td>0.109311</td>\n",
       "      <td>0.029308</td>\n",
       "      <td>0.029044</td>\n",
       "      <td>0.107101</td>\n",
       "      <td>-0.007541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.151611</td>\n",
       "      <td>0.148789</td>\n",
       "      <td>0.090900</td>\n",
       "      <td>-0.000616</td>\n",
       "      <td>0.151533</td>\n",
       "      <td>0.079336</td>\n",
       "      <td>-0.004712</td>\n",
       "      <td>-0.077400</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>0.031200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bA9F11.1</th>\n",
       "      <td>0.150150</td>\n",
       "      <td>0.095300</td>\n",
       "      <td>-0.018000</td>\n",
       "      <td>0.263000</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>-0.021791</td>\n",
       "      <td>0.147900</td>\n",
       "      <td>-0.086950</td>\n",
       "      <td>-0.105000</td>\n",
       "      <td>-0.064150</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.048850</td>\n",
       "      <td>0.065850</td>\n",
       "      <td>0.111600</td>\n",
       "      <td>-0.127850</td>\n",
       "      <td>-0.217500</td>\n",
       "      <td>-0.151800</td>\n",
       "      <td>-0.283000</td>\n",
       "      <td>-0.072160</td>\n",
       "      <td>0.191000</td>\n",
       "      <td>0.003150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <td>0.006610</td>\n",
       "      <td>-0.161000</td>\n",
       "      <td>-0.120000</td>\n",
       "      <td>-0.133000</td>\n",
       "      <td>-0.138000</td>\n",
       "      <td>-0.177000</td>\n",
       "      <td>-0.132000</td>\n",
       "      <td>0.051200</td>\n",
       "      <td>-0.039200</td>\n",
       "      <td>-0.014300</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.188000</td>\n",
       "      <td>-0.012400</td>\n",
       "      <td>0.119000</td>\n",
       "      <td>-0.193000</td>\n",
       "      <td>-0.092300</td>\n",
       "      <td>-0.124000</td>\n",
       "      <td>-0.271000</td>\n",
       "      <td>-0.176000</td>\n",
       "      <td>-0.100000</td>\n",
       "      <td>-0.269000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-D3</th>\n",
       "      <td>-0.104746</td>\n",
       "      <td>-0.010522</td>\n",
       "      <td>0.026106</td>\n",
       "      <td>0.072607</td>\n",
       "      <td>-0.014310</td>\n",
       "      <td>0.027170</td>\n",
       "      <td>0.043281</td>\n",
       "      <td>0.107378</td>\n",
       "      <td>0.029616</td>\n",
       "      <td>-0.188333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.173267</td>\n",
       "      <td>-0.299222</td>\n",
       "      <td>-0.090567</td>\n",
       "      <td>-0.004926</td>\n",
       "      <td>0.055044</td>\n",
       "      <td>0.054733</td>\n",
       "      <td>0.012011</td>\n",
       "      <td>0.007011</td>\n",
       "      <td>0.017700</td>\n",
       "      <td>0.026992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-H2</th>\n",
       "      <td>0.053800</td>\n",
       "      <td>0.027662</td>\n",
       "      <td>0.011612</td>\n",
       "      <td>0.063725</td>\n",
       "      <td>-0.114200</td>\n",
       "      <td>0.026400</td>\n",
       "      <td>0.140775</td>\n",
       "      <td>-0.134775</td>\n",
       "      <td>-0.026850</td>\n",
       "      <td>0.035800</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.031643</td>\n",
       "      <td>-0.244000</td>\n",
       "      <td>0.083365</td>\n",
       "      <td>-0.079825</td>\n",
       "      <td>-0.076025</td>\n",
       "      <td>-0.071883</td>\n",
       "      <td>-0.031462</td>\n",
       "      <td>0.068550</td>\n",
       "      <td>0.098075</td>\n",
       "      <td>0.060400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mimitin</th>\n",
       "      <td>-0.052605</td>\n",
       "      <td>0.082836</td>\n",
       "      <td>0.032056</td>\n",
       "      <td>-0.053923</td>\n",
       "      <td>-0.005092</td>\n",
       "      <td>-0.007739</td>\n",
       "      <td>-0.010476</td>\n",
       "      <td>0.012062</td>\n",
       "      <td>-0.002436</td>\n",
       "      <td>-0.199143</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130712</td>\n",
       "      <td>0.100208</td>\n",
       "      <td>0.085496</td>\n",
       "      <td>0.004645</td>\n",
       "      <td>0.171510</td>\n",
       "      <td>-0.019704</td>\n",
       "      <td>0.012892</td>\n",
       "      <td>-0.295004</td>\n",
       "      <td>0.053736</td>\n",
       "      <td>-0.030301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16832 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "Gene                                                                           \n",
       "            -0.091366  -0.009954   0.064026   0.223229   0.015514   0.133943   \n",
       "15E1.2      -0.018767  -0.036233   0.126533   0.015967  -0.001400  -0.229000   \n",
       "2'-PDE      -0.095650  -0.038200  -0.273500  -0.188500   0.077200  -0.170500   \n",
       "76P         -0.177225  -0.016975  -0.015500  -0.230250  -0.381500  -0.132625   \n",
       "7A5         -0.060111   0.072104   0.055131  -0.020300   0.201556   0.109311   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "bA9F11.1     0.150150   0.095300  -0.018000   0.263000   0.034750  -0.021791   \n",
       "dJ341D10.1   0.006610  -0.161000  -0.120000  -0.133000  -0.138000  -0.177000   \n",
       "hCAP-D3     -0.104746  -0.010522   0.026106   0.072607  -0.014310   0.027170   \n",
       "hCAP-H2      0.053800   0.027662   0.011612   0.063725  -0.114200   0.026400   \n",
       "mimitin     -0.052605   0.082836   0.032056  -0.053923  -0.005092  -0.007739   \n",
       "\n",
       "            GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  \\\n",
       "Gene                                                    ...              \n",
       "             0.010000   0.036996  -0.109300   0.058823  ...   0.152000   \n",
       "15E1.2       0.095173  -0.137293  -0.145300   0.137067  ...   0.146700   \n",
       "2'-PDE      -0.023250  -0.265000  -0.224050   0.028700  ...  -0.189500   \n",
       "76P         -0.074625   0.013000  -0.158050  -0.093265  ...   0.067575   \n",
       "7A5          0.029308   0.029044   0.107101  -0.007541  ...   0.151611   \n",
       "...               ...        ...        ...        ...  ...        ...   \n",
       "bA9F11.1     0.147900  -0.086950  -0.105000  -0.064150  ...  -0.048850   \n",
       "dJ341D10.1  -0.132000   0.051200  -0.039200  -0.014300  ...  -0.188000   \n",
       "hCAP-D3      0.043281   0.107378   0.029616  -0.188333  ...   0.173267   \n",
       "hCAP-H2      0.140775  -0.134775  -0.026850   0.035800  ...  -0.031643   \n",
       "mimitin     -0.010476   0.012062  -0.002436  -0.199143  ...   0.130712   \n",
       "\n",
       "            GSM540602  GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  \\\n",
       "Gene                                                                           \n",
       "             0.040586   0.114029  -0.021514   0.110771   0.135557   0.125829   \n",
       "15E1.2       0.267000   0.207333   0.039400  -0.047812   0.011360  -0.109767   \n",
       "2'-PDE      -0.117450  -0.166800  -0.037500  -0.123500   0.019000  -0.111050   \n",
       "76P          0.068450  -0.156500   0.048800  -0.054375  -0.010340  -0.110975   \n",
       "7A5          0.148789   0.090900  -0.000616   0.151533   0.079336  -0.004712   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "bA9F11.1     0.065850   0.111600  -0.127850  -0.217500  -0.151800  -0.283000   \n",
       "dJ341D10.1  -0.012400   0.119000  -0.193000  -0.092300  -0.124000  -0.271000   \n",
       "hCAP-D3     -0.299222  -0.090567  -0.004926   0.055044   0.054733   0.012011   \n",
       "hCAP-H2     -0.244000   0.083365  -0.079825  -0.076025  -0.071883  -0.031462   \n",
       "mimitin      0.100208   0.085496   0.004645   0.171510  -0.019704   0.012892   \n",
       "\n",
       "            GSM540608  GSM540609  GSM540610  \n",
       "Gene                                         \n",
       "             0.009514   0.174671   0.049243  \n",
       "15E1.2       0.300333   0.013167  -0.008633  \n",
       "2'-PDE       0.182000  -0.032200  -0.052400  \n",
       "76P          0.131200  -0.107950  -0.035025  \n",
       "7A5         -0.077400  -0.000033   0.031200  \n",
       "...               ...        ...        ...  \n",
       "bA9F11.1    -0.072160   0.191000   0.003150  \n",
       "dJ341D10.1  -0.176000  -0.100000  -0.269000  \n",
       "hCAP-D3      0.007011   0.017700   0.026992  \n",
       "hCAP-H2      0.068550   0.098075   0.060400  \n",
       "mimitin     -0.295004   0.053736  -0.030301  \n",
       "\n",
       "[16832 rows x 61 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.261000</td>\n",
       "      <td>0.023700</td>\n",
       "      <td>0.117000</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>0.141000</td>\n",
       "      <td>0.239000</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>0.216000</td>\n",
       "      <td>0.318000</td>\n",
       "      <td>0.085300</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.164000</td>\n",
       "      <td>0.109000</td>\n",
       "      <td>0.205000</td>\n",
       "      <td>-0.053900</td>\n",
       "      <td>-0.069000</td>\n",
       "      <td>-0.034000</td>\n",
       "      <td>0.054000</td>\n",
       "      <td>-0.154000</td>\n",
       "      <td>0.242000</td>\n",
       "      <td>0.074100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-0.071700</td>\n",
       "      <td>-0.075717</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>0.176317</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>0.049178</td>\n",
       "      <td>-0.039567</td>\n",
       "      <td>0.045017</td>\n",
       "      <td>0.107967</td>\n",
       "      <td>0.116025</td>\n",
       "      <td>...</td>\n",
       "      <td>0.539667</td>\n",
       "      <td>0.245167</td>\n",
       "      <td>0.143050</td>\n",
       "      <td>0.039052</td>\n",
       "      <td>-0.059468</td>\n",
       "      <td>0.115183</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>0.168233</td>\n",
       "      <td>0.139317</td>\n",
       "      <td>0.074317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.056388</td>\n",
       "      <td>-0.025460</td>\n",
       "      <td>0.121440</td>\n",
       "      <td>0.166813</td>\n",
       "      <td>0.002068</td>\n",
       "      <td>0.105455</td>\n",
       "      <td>0.087587</td>\n",
       "      <td>0.114687</td>\n",
       "      <td>0.150363</td>\n",
       "      <td>0.049269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.537412</td>\n",
       "      <td>0.228625</td>\n",
       "      <td>0.226325</td>\n",
       "      <td>0.063661</td>\n",
       "      <td>0.065370</td>\n",
       "      <td>-0.013030</td>\n",
       "      <td>-0.029762</td>\n",
       "      <td>0.093887</td>\n",
       "      <td>0.101675</td>\n",
       "      <td>-0.023225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.115900</td>\n",
       "      <td>0.056557</td>\n",
       "      <td>-0.057271</td>\n",
       "      <td>0.139629</td>\n",
       "      <td>-0.060130</td>\n",
       "      <td>-0.018657</td>\n",
       "      <td>0.206329</td>\n",
       "      <td>-0.027757</td>\n",
       "      <td>-0.015286</td>\n",
       "      <td>-0.037541</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.090871</td>\n",
       "      <td>-0.217963</td>\n",
       "      <td>-0.032471</td>\n",
       "      <td>-0.165701</td>\n",
       "      <td>-0.124071</td>\n",
       "      <td>-0.120486</td>\n",
       "      <td>-0.127914</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>0.150286</td>\n",
       "      <td>-0.038199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>0.185500</td>\n",
       "      <td>-0.096250</td>\n",
       "      <td>-0.289000</td>\n",
       "      <td>-0.030060</td>\n",
       "      <td>-0.063700</td>\n",
       "      <td>-0.112750</td>\n",
       "      <td>-0.056290</td>\n",
       "      <td>0.054900</td>\n",
       "      <td>0.011600</td>\n",
       "      <td>0.062265</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.253000</td>\n",
       "      <td>0.114100</td>\n",
       "      <td>-0.105850</td>\n",
       "      <td>-0.118550</td>\n",
       "      <td>-0.061350</td>\n",
       "      <td>0.013800</td>\n",
       "      <td>0.006250</td>\n",
       "      <td>-0.114600</td>\n",
       "      <td>-0.029010</td>\n",
       "      <td>-0.064800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>0.021000</td>\n",
       "      <td>-0.025660</td>\n",
       "      <td>-0.214200</td>\n",
       "      <td>-0.152360</td>\n",
       "      <td>-0.020380</td>\n",
       "      <td>-0.004088</td>\n",
       "      <td>-0.022060</td>\n",
       "      <td>0.012460</td>\n",
       "      <td>0.035980</td>\n",
       "      <td>-0.095848</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.150060</td>\n",
       "      <td>-0.114840</td>\n",
       "      <td>-0.113584</td>\n",
       "      <td>-0.069040</td>\n",
       "      <td>-0.112720</td>\n",
       "      <td>-0.080224</td>\n",
       "      <td>-0.191920</td>\n",
       "      <td>-0.193660</td>\n",
       "      <td>-0.168400</td>\n",
       "      <td>-0.114120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.018978</td>\n",
       "      <td>-0.048632</td>\n",
       "      <td>-0.240545</td>\n",
       "      <td>-0.211436</td>\n",
       "      <td>-0.014087</td>\n",
       "      <td>-0.029066</td>\n",
       "      <td>0.012171</td>\n",
       "      <td>0.010093</td>\n",
       "      <td>0.011982</td>\n",
       "      <td>-0.161527</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.117764</td>\n",
       "      <td>-0.096468</td>\n",
       "      <td>-0.157382</td>\n",
       "      <td>-0.115042</td>\n",
       "      <td>-0.050454</td>\n",
       "      <td>-0.110373</td>\n",
       "      <td>-0.230038</td>\n",
       "      <td>-0.178318</td>\n",
       "      <td>-0.209545</td>\n",
       "      <td>-0.195591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>-0.034700</td>\n",
       "      <td>0.021967</td>\n",
       "      <td>0.089867</td>\n",
       "      <td>0.084667</td>\n",
       "      <td>-0.098300</td>\n",
       "      <td>0.109900</td>\n",
       "      <td>0.071067</td>\n",
       "      <td>-0.106867</td>\n",
       "      <td>-0.011533</td>\n",
       "      <td>0.158667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.127000</td>\n",
       "      <td>0.010613</td>\n",
       "      <td>0.163800</td>\n",
       "      <td>-0.020167</td>\n",
       "      <td>-0.006195</td>\n",
       "      <td>0.057200</td>\n",
       "      <td>0.041467</td>\n",
       "      <td>0.072733</td>\n",
       "      <td>0.206333</td>\n",
       "      <td>0.196667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>-0.129433</td>\n",
       "      <td>-0.158727</td>\n",
       "      <td>-0.175639</td>\n",
       "      <td>-0.123181</td>\n",
       "      <td>-0.021133</td>\n",
       "      <td>-0.288200</td>\n",
       "      <td>-0.128013</td>\n",
       "      <td>-0.138720</td>\n",
       "      <td>-0.097573</td>\n",
       "      <td>0.044883</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.167607</td>\n",
       "      <td>-0.057523</td>\n",
       "      <td>-0.091887</td>\n",
       "      <td>0.025984</td>\n",
       "      <td>-0.078663</td>\n",
       "      <td>-0.026909</td>\n",
       "      <td>-0.100020</td>\n",
       "      <td>0.083072</td>\n",
       "      <td>-0.234480</td>\n",
       "      <td>-0.077160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>-0.079166</td>\n",
       "      <td>-0.043763</td>\n",
       "      <td>-0.262375</td>\n",
       "      <td>-0.159617</td>\n",
       "      <td>0.018503</td>\n",
       "      <td>-0.025875</td>\n",
       "      <td>-0.000616</td>\n",
       "      <td>0.043308</td>\n",
       "      <td>0.033308</td>\n",
       "      <td>-0.232042</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.136790</td>\n",
       "      <td>-0.071925</td>\n",
       "      <td>-0.230033</td>\n",
       "      <td>-0.103192</td>\n",
       "      <td>0.032115</td>\n",
       "      <td>-0.087575</td>\n",
       "      <td>-0.126593</td>\n",
       "      <td>-0.300383</td>\n",
       "      <td>-0.298500</td>\n",
       "      <td>-0.095075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12669 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "A1BG     0.261000   0.023700   0.117000   0.154000   0.141000   0.239000   \n",
       "A2M     -0.071700  -0.075717   0.184000   0.176317   0.034750   0.049178   \n",
       "A2ML1   -0.056388  -0.025460   0.121440   0.166813   0.002068   0.105455   \n",
       "A4GALT   0.115900   0.056557  -0.057271   0.139629  -0.060130  -0.018657   \n",
       "A4GNT    0.185500  -0.096250  -0.289000  -0.030060  -0.063700  -0.112750   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A   0.021000  -0.025660  -0.214200  -0.152360  -0.020380  -0.004088   \n",
       "ZYG11B   0.018978  -0.048632  -0.240545  -0.211436  -0.014087  -0.029066   \n",
       "ZYX     -0.034700   0.021967   0.089867   0.084667  -0.098300   0.109900   \n",
       "ZZEF1   -0.129433  -0.158727  -0.175639  -0.123181  -0.021133  -0.288200   \n",
       "ZZZ3    -0.079166  -0.043763  -0.262375  -0.159617   0.018503  -0.025875   \n",
       "\n",
       "        GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  GSM540602  \\\n",
       "A1BG     0.197000   0.216000   0.318000   0.085300  ...  -0.164000   0.109000   \n",
       "A2M     -0.039567   0.045017   0.107967   0.116025  ...   0.539667   0.245167   \n",
       "A2ML1    0.087587   0.114687   0.150363   0.049269  ...   0.537412   0.228625   \n",
       "A4GALT   0.206329  -0.027757  -0.015286  -0.037541  ...  -0.090871  -0.217963   \n",
       "A4GNT   -0.056290   0.054900   0.011600   0.062265  ...  -0.253000   0.114100   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "ZYG11A  -0.022060   0.012460   0.035980  -0.095848  ...  -0.150060  -0.114840   \n",
       "ZYG11B   0.012171   0.010093   0.011982  -0.161527  ...  -0.117764  -0.096468   \n",
       "ZYX      0.071067  -0.106867  -0.011533   0.158667  ...   0.127000   0.010613   \n",
       "ZZEF1   -0.128013  -0.138720  -0.097573   0.044883  ...  -0.167607  -0.057523   \n",
       "ZZZ3    -0.000616   0.043308   0.033308  -0.232042  ...  -0.136790  -0.071925   \n",
       "\n",
       "        GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  GSM540608  \\\n",
       "A1BG     0.205000  -0.053900  -0.069000  -0.034000   0.054000  -0.154000   \n",
       "A2M      0.143050   0.039052  -0.059468   0.115183   0.112083   0.168233   \n",
       "A2ML1    0.226325   0.063661   0.065370  -0.013030  -0.029762   0.093887   \n",
       "A4GALT  -0.032471  -0.165701  -0.124071  -0.120486  -0.127914   0.012549   \n",
       "A4GNT   -0.105850  -0.118550  -0.061350   0.013800   0.006250  -0.114600   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A  -0.113584  -0.069040  -0.112720  -0.080224  -0.191920  -0.193660   \n",
       "ZYG11B  -0.157382  -0.115042  -0.050454  -0.110373  -0.230038  -0.178318   \n",
       "ZYX      0.163800  -0.020167  -0.006195   0.057200   0.041467   0.072733   \n",
       "ZZEF1   -0.091887   0.025984  -0.078663  -0.026909  -0.100020   0.083072   \n",
       "ZZZ3    -0.230033  -0.103192   0.032115  -0.087575  -0.126593  -0.300383   \n",
       "\n",
       "        GSM540609  GSM540610  \n",
       "A1BG     0.242000   0.074100  \n",
       "A2M      0.139317   0.074317  \n",
       "A2ML1    0.101675  -0.023225  \n",
       "A4GALT   0.150286  -0.038199  \n",
       "A4GNT   -0.029010  -0.064800  \n",
       "...           ...        ...  \n",
       "ZYG11A  -0.168400  -0.114120  \n",
       "ZYG11B  -0.209545  -0.195591  \n",
       "ZYX      0.206333   0.196667  \n",
       "ZZEF1   -0.234480  -0.077160  \n",
       "ZZZ3    -0.298500  -0.095075  \n",
       "\n",
       "[12669 rows x 61 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM540550</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.071700</td>\n",
       "      <td>-0.056388</td>\n",
       "      <td>0.115900</td>\n",
       "      <td>0.18550</td>\n",
       "      <td>-0.074332</td>\n",
       "      <td>0.284667</td>\n",
       "      <td>-0.050671</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011806</td>\n",
       "      <td>-0.069583</td>\n",
       "      <td>0.0818</td>\n",
       "      <td>-0.26000</td>\n",
       "      <td>0.19156</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>0.018978</td>\n",
       "      <td>-0.034700</td>\n",
       "      <td>-0.129433</td>\n",
       "      <td>-0.079166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540551</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0237</td>\n",
       "      <td>-0.075717</td>\n",
       "      <td>-0.025460</td>\n",
       "      <td>0.056557</td>\n",
       "      <td>-0.09625</td>\n",
       "      <td>0.058537</td>\n",
       "      <td>0.048833</td>\n",
       "      <td>-0.124633</td>\n",
       "      <td>-0.00876</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074980</td>\n",
       "      <td>0.105617</td>\n",
       "      <td>0.1890</td>\n",
       "      <td>-0.39000</td>\n",
       "      <td>-0.02402</td>\n",
       "      <td>-0.025660</td>\n",
       "      <td>-0.048632</td>\n",
       "      <td>0.021967</td>\n",
       "      <td>-0.158727</td>\n",
       "      <td>-0.043763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540552</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>0.121440</td>\n",
       "      <td>-0.057271</td>\n",
       "      <td>-0.28900</td>\n",
       "      <td>0.049406</td>\n",
       "      <td>0.168933</td>\n",
       "      <td>0.118744</td>\n",
       "      <td>-0.20600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117400</td>\n",
       "      <td>0.014768</td>\n",
       "      <td>0.1560</td>\n",
       "      <td>-0.51600</td>\n",
       "      <td>-0.26312</td>\n",
       "      <td>-0.214200</td>\n",
       "      <td>-0.240545</td>\n",
       "      <td>0.089867</td>\n",
       "      <td>-0.175639</td>\n",
       "      <td>-0.262375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540553</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1540</td>\n",
       "      <td>0.176317</td>\n",
       "      <td>0.166813</td>\n",
       "      <td>0.139629</td>\n",
       "      <td>-0.03006</td>\n",
       "      <td>-0.036533</td>\n",
       "      <td>0.314667</td>\n",
       "      <td>0.071756</td>\n",
       "      <td>0.12700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.120940</td>\n",
       "      <td>0.090387</td>\n",
       "      <td>0.0823</td>\n",
       "      <td>-0.42100</td>\n",
       "      <td>0.12542</td>\n",
       "      <td>-0.152360</td>\n",
       "      <td>-0.211436</td>\n",
       "      <td>0.084667</td>\n",
       "      <td>-0.123181</td>\n",
       "      <td>-0.159617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540554</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1410</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>0.002068</td>\n",
       "      <td>-0.060130</td>\n",
       "      <td>-0.06370</td>\n",
       "      <td>0.322153</td>\n",
       "      <td>0.049267</td>\n",
       "      <td>-0.056361</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015800</td>\n",
       "      <td>0.022417</td>\n",
       "      <td>0.0821</td>\n",
       "      <td>-0.58000</td>\n",
       "      <td>0.07364</td>\n",
       "      <td>-0.020380</td>\n",
       "      <td>-0.014087</td>\n",
       "      <td>-0.098300</td>\n",
       "      <td>-0.021133</td>\n",
       "      <td>0.018503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540606</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.0340</td>\n",
       "      <td>0.115183</td>\n",
       "      <td>-0.013030</td>\n",
       "      <td>-0.120486</td>\n",
       "      <td>0.01380</td>\n",
       "      <td>0.096020</td>\n",
       "      <td>-0.082280</td>\n",
       "      <td>-0.084410</td>\n",
       "      <td>0.14000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053585</td>\n",
       "      <td>0.004537</td>\n",
       "      <td>0.0128</td>\n",
       "      <td>0.00522</td>\n",
       "      <td>-0.05314</td>\n",
       "      <td>-0.080224</td>\n",
       "      <td>-0.110373</td>\n",
       "      <td>0.057200</td>\n",
       "      <td>-0.026909</td>\n",
       "      <td>-0.087575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540607</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>-0.029762</td>\n",
       "      <td>-0.127914</td>\n",
       "      <td>0.00625</td>\n",
       "      <td>0.126520</td>\n",
       "      <td>-0.074100</td>\n",
       "      <td>-0.090500</td>\n",
       "      <td>0.15000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000366</td>\n",
       "      <td>-0.128767</td>\n",
       "      <td>0.0391</td>\n",
       "      <td>0.53800</td>\n",
       "      <td>-0.00796</td>\n",
       "      <td>-0.191920</td>\n",
       "      <td>-0.230038</td>\n",
       "      <td>0.041467</td>\n",
       "      <td>-0.100020</td>\n",
       "      <td>-0.126593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540608</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.1540</td>\n",
       "      <td>0.168233</td>\n",
       "      <td>0.093887</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>-0.11460</td>\n",
       "      <td>-0.061250</td>\n",
       "      <td>0.163033</td>\n",
       "      <td>0.201889</td>\n",
       "      <td>0.07380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.075702</td>\n",
       "      <td>0.154717</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.50600</td>\n",
       "      <td>0.14348</td>\n",
       "      <td>-0.193660</td>\n",
       "      <td>-0.178318</td>\n",
       "      <td>0.072733</td>\n",
       "      <td>0.083072</td>\n",
       "      <td>-0.300383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540609</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2420</td>\n",
       "      <td>0.139317</td>\n",
       "      <td>0.101675</td>\n",
       "      <td>0.150286</td>\n",
       "      <td>-0.02901</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>0.065833</td>\n",
       "      <td>-0.030744</td>\n",
       "      <td>0.07190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.066480</td>\n",
       "      <td>-0.007978</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>-0.48600</td>\n",
       "      <td>0.05238</td>\n",
       "      <td>-0.168400</td>\n",
       "      <td>-0.209545</td>\n",
       "      <td>0.206333</td>\n",
       "      <td>-0.234480</td>\n",
       "      <td>-0.298500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540610</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0741</td>\n",
       "      <td>0.074317</td>\n",
       "      <td>-0.023225</td>\n",
       "      <td>-0.038199</td>\n",
       "      <td>-0.06480</td>\n",
       "      <td>-0.003480</td>\n",
       "      <td>0.033500</td>\n",
       "      <td>-0.013467</td>\n",
       "      <td>0.08060</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030380</td>\n",
       "      <td>-0.102450</td>\n",
       "      <td>0.0841</td>\n",
       "      <td>-0.44800</td>\n",
       "      <td>0.11162</td>\n",
       "      <td>-0.114120</td>\n",
       "      <td>-0.195591</td>\n",
       "      <td>0.196667</td>\n",
       "      <td>-0.077160</td>\n",
       "      <td>-0.095075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 12670 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer    A1BG       A2M     A2ML1    A4GALT  \\\n",
       "GSM540550                    0.0  0.2610 -0.071700 -0.056388  0.115900   \n",
       "GSM540551                    0.0  0.0237 -0.075717 -0.025460  0.056557   \n",
       "GSM540552                    0.0  0.1170  0.184000  0.121440 -0.057271   \n",
       "GSM540553                    0.0  0.1540  0.176317  0.166813  0.139629   \n",
       "GSM540554                    0.0  0.1410  0.034750  0.002068 -0.060130   \n",
       "...                          ...     ...       ...       ...       ...   \n",
       "GSM540606                    0.0 -0.0340  0.115183 -0.013030 -0.120486   \n",
       "GSM540607                    0.0  0.0540  0.112083 -0.029762 -0.127914   \n",
       "GSM540608                    0.0 -0.1540  0.168233  0.093887  0.012549   \n",
       "GSM540609                    0.0  0.2420  0.139317  0.101675  0.150286   \n",
       "GSM540610                    0.0  0.0741  0.074317 -0.023225 -0.038199   \n",
       "\n",
       "             A4GNT      AAA1      AAAS      AACS    AADAC  ...      ZW10  \\\n",
       "GSM540550  0.18550 -0.074332  0.284667 -0.050671  0.14100  ... -0.011806   \n",
       "GSM540551 -0.09625  0.058537  0.048833 -0.124633 -0.00876  ...  0.074980   \n",
       "GSM540552 -0.28900  0.049406  0.168933  0.118744 -0.20600  ...  0.117400   \n",
       "GSM540553 -0.03006 -0.036533  0.314667  0.071756  0.12700  ...  0.120940   \n",
       "GSM540554 -0.06370  0.322153  0.049267 -0.056361  0.14400  ...  0.015800   \n",
       "...            ...       ...       ...       ...      ...  ...       ...   \n",
       "GSM540606  0.01380  0.096020 -0.082280 -0.084410  0.14000  ...  0.053585   \n",
       "GSM540607  0.00625  0.126520 -0.074100 -0.090500  0.15000  ... -0.000366   \n",
       "GSM540608 -0.11460 -0.061250  0.163033  0.201889  0.07380  ...  0.075702   \n",
       "GSM540609 -0.02901  0.000785  0.065833 -0.030744  0.07190  ...  0.066480   \n",
       "GSM540610 -0.06480 -0.003480  0.033500 -0.013467  0.08060  ... -0.030380   \n",
       "\n",
       "             ZWILCH   ZWINT     ZXDB     ZXDC    ZYG11A    ZYG11B       ZYX  \\\n",
       "GSM540550 -0.069583  0.0818 -0.26000  0.19156  0.021000  0.018978 -0.034700   \n",
       "GSM540551  0.105617  0.1890 -0.39000 -0.02402 -0.025660 -0.048632  0.021967   \n",
       "GSM540552  0.014768  0.1560 -0.51600 -0.26312 -0.214200 -0.240545  0.089867   \n",
       "GSM540553  0.090387  0.0823 -0.42100  0.12542 -0.152360 -0.211436  0.084667   \n",
       "GSM540554  0.022417  0.0821 -0.58000  0.07364 -0.020380 -0.014087 -0.098300   \n",
       "...             ...     ...      ...      ...       ...       ...       ...   \n",
       "GSM540606  0.004537  0.0128  0.00522 -0.05314 -0.080224 -0.110373  0.057200   \n",
       "GSM540607 -0.128767  0.0391  0.53800 -0.00796 -0.191920 -0.230038  0.041467   \n",
       "GSM540608  0.154717  0.1470 -0.50600  0.14348 -0.193660 -0.178318  0.072733   \n",
       "GSM540609 -0.007978  0.1480 -0.48600  0.05238 -0.168400 -0.209545  0.206333   \n",
       "GSM540610 -0.102450  0.0841 -0.44800  0.11162 -0.114120 -0.195591  0.196667   \n",
       "\n",
       "              ZZEF1      ZZZ3  \n",
       "GSM540550 -0.129433 -0.079166  \n",
       "GSM540551 -0.158727 -0.043763  \n",
       "GSM540552 -0.175639 -0.262375  \n",
       "GSM540553 -0.123181 -0.159617  \n",
       "GSM540554 -0.021133  0.018503  \n",
       "...             ...       ...  \n",
       "GSM540606 -0.026909 -0.087575  \n",
       "GSM540607 -0.100020 -0.126593  \n",
       "GSM540608  0.083072 -0.300383  \n",
       "GSM540609 -0.234480 -0.298500  \n",
       "GSM540610 -0.077160 -0.095075  \n",
       "\n",
       "[61 rows x 12670 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 61 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 25 occurrences. This represents 40.98% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Advancing a Clinically Relevant Perspective of the Clonal Nature of Cancer\"\n",
      "!Series_summary\t\"We used DNA content-based flow cytometry to distinguish and isolate nuclei from clonal populations in primary tissues from three disparate cancers with variable clinical histories. We then developed a methodology to adapt flow cytometrically purified nuclei samples for use with whole genome technologies including aCGH and next generation sequencing (NGS). Our results demonstrate that selected aberrations in the genomes of distinct clonal populations in each patient create clinically relevant contexts at least with respect to the cancer types profiled in this study.\"\n",
      "!Series_overall_design\t\"We applied DNA content based flow sorting to isolate the nuclei of clonal populations from tumor biopsies. Genomic DNA from each sorted population was amplified with phi29 polymerase. A 1ug aliquot of each amplified sample was digested with DNAse 1 then labeled with Cy5 using a Klenow-based commercial kit (Invitrogen). Each sample was hybridized with a pooled normal (46,XX) reference (Promega) to Agilent 244k CGH arrays. The use of highly purified objectively defined flow sorted populations provides high definition genomic profiles of clonal populations from pancreatic adenocarcinomas (PA), adrenal cortical carcinomas (ACC), and prostate adenocarcinomas (PC).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                 GSM540550  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540551  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540552  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540553  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540554  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540555  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540556  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540557  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540558  ...  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma  ...   \n",
       "\n",
       "                            GSM540601                           GSM540602  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Adrenal Cortical Carcinoma   \n",
       "\n",
       "                            GSM540603                   GSM540604  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540605                   GSM540606  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540607                   GSM540608  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540609                   GSM540610  \n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "\n",
    "cohort = accession_num = \"GSE21660\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Adrenal Cortical Carcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "\n",
       "                       GSM540555  GSM540556  GSM540557  GSM540558  GSM540559  \\\n",
       "Adrenocortical Cancer          1          1          1          1          1   \n",
       "\n",
       "                       ...  GSM540601  GSM540602  GSM540603  GSM540604  \\\n",
       "Adrenocortical Cancer  ...          0          0          0          0   \n",
       "\n",
       "                       GSM540605  GSM540606  GSM540607  GSM540608  GSM540609  \\\n",
       "Adrenocortical Cancer          0          0          0          0          0   \n",
       "\n",
       "                       GSM540610  \n",
       "Adrenocortical Cancer          0  \n",
       "\n",
       "[1 rows x 61 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.00599</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>-0.0779</td>\n",
       "      <td>0.01020</td>\n",
       "      <td>-0.00558</td>\n",
       "      <td>-0.0871</td>\n",
       "      <td>-0.0895</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.00695</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0563</td>\n",
       "      <td>0.002070</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.00841</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.01280</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>-0.000229</td>\n",
       "      <td>-0.0148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1730</td>\n",
       "      <td>-0.43800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.6140</td>\n",
       "      <td>-0.0190</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.5670</td>\n",
       "      <td>-0.004620</td>\n",
       "      <td>-0.4700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.2000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3190</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.4420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.6850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.19900</td>\n",
       "      <td>0.1570</td>\n",
       "      <td>0.2230</td>\n",
       "      <td>0.01230</td>\n",
       "      <td>0.13200</td>\n",
       "      <td>0.0682</td>\n",
       "      <td>0.1330</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>0.03560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1920</td>\n",
       "      <td>0.3650</td>\n",
       "      <td>0.364000</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>-0.24800</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>-0.08650</td>\n",
       "      <td>0.0336</td>\n",
       "      <td>0.167000</td>\n",
       "      <td>0.0337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.1050</td>\n",
       "      <td>-0.06010</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.1780</td>\n",
       "      <td>-0.00527</td>\n",
       "      <td>0.00652</td>\n",
       "      <td>-0.1610</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>-0.1520</td>\n",
       "      <td>-0.02830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0991</td>\n",
       "      <td>-0.0738</td>\n",
       "      <td>-0.058400</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>-0.06350</td>\n",
       "      <td>-0.1470</td>\n",
       "      <td>-0.24700</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>-0.115000</td>\n",
       "      <td>0.0135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243500</th>\n",
       "      <td>-0.2810</td>\n",
       "      <td>-0.34800</td>\n",
       "      <td>-0.3390</td>\n",
       "      <td>-0.2320</td>\n",
       "      <td>0.02470</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>-0.2440</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>-0.1790</td>\n",
       "      <td>-0.34600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>-0.1520</td>\n",
       "      <td>-0.211000</td>\n",
       "      <td>-0.0559</td>\n",
       "      <td>0.08870</td>\n",
       "      <td>-0.0608</td>\n",
       "      <td>0.00423</td>\n",
       "      <td>-0.4000</td>\n",
       "      <td>-0.069500</td>\n",
       "      <td>-0.0305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243501</th>\n",
       "      <td>-0.1770</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.2670</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.28800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0433</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2160</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.122000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243502</th>\n",
       "      <td>-0.5300</td>\n",
       "      <td>1.30000</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.66700</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.2540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.1480</td>\n",
       "      <td>-0.302000</td>\n",
       "      <td>-0.9700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243503</th>\n",
       "      <td>0.0344</td>\n",
       "      <td>0.02240</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>-0.0692</td>\n",
       "      <td>0.03500</td>\n",
       "      <td>0.01120</td>\n",
       "      <td>-0.0938</td>\n",
       "      <td>-0.0330</td>\n",
       "      <td>-0.1030</td>\n",
       "      <td>0.00613</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0709</td>\n",
       "      <td>0.0626</td>\n",
       "      <td>-0.000328</td>\n",
       "      <td>0.0182</td>\n",
       "      <td>0.01140</td>\n",
       "      <td>0.0342</td>\n",
       "      <td>-0.03030</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>-0.0138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243504</th>\n",
       "      <td>0.0434</td>\n",
       "      <td>0.01580</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>0.05200</td>\n",
       "      <td>0.00951</td>\n",
       "      <td>-0.0797</td>\n",
       "      <td>-0.0113</td>\n",
       "      <td>-0.1080</td>\n",
       "      <td>-0.00240</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0639</td>\n",
       "      <td>0.0483</td>\n",
       "      <td>-0.002440</td>\n",
       "      <td>0.0539</td>\n",
       "      <td>0.02960</td>\n",
       "      <td>0.0264</td>\n",
       "      <td>-0.00543</td>\n",
       "      <td>0.0700</td>\n",
       "      <td>-0.012400</td>\n",
       "      <td>-0.0203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243429 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "ID                                                                         \n",
       "1          0.0201    0.00599     0.0117    -0.0779    0.01020   -0.00558   \n",
       "2          0.0000    0.00000     0.0000    -0.1730   -0.43800    0.00000   \n",
       "3         -0.2000    0.00000     0.0000     0.0000    0.00000    0.00000   \n",
       "4          0.1210    0.19900     0.1570     0.2230    0.01230    0.13200   \n",
       "5         -0.1050   -0.06010    -0.1140    -0.1780   -0.00527    0.00652   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "243500    -0.2810   -0.34800    -0.3390    -0.2320    0.02470    0.02690   \n",
       "243501    -0.1770    0.00000    -0.2670     0.0000   -0.28800    0.00000   \n",
       "243502    -0.5300    1.30000     0.0179     0.0000   -0.66700    0.00000   \n",
       "243503     0.0344    0.02240     0.0275    -0.0692    0.03500    0.01120   \n",
       "243504     0.0434    0.01580     0.0122    -0.0646    0.05200    0.00951   \n",
       "\n",
       "        GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  GSM540602  \\\n",
       "ID                                                  ...                         \n",
       "1         -0.0871    -0.0895    -0.1140   -0.00695  ...     0.0592     0.0563   \n",
       "2          0.0000    -0.6140    -0.0190    0.00000  ...     0.0000     0.0000   \n",
       "3          0.0000    -0.3190     0.0000    0.00000  ...     0.0000     0.0000   \n",
       "4          0.0682     0.1330     0.1480    0.03560  ...     0.1920     0.3650   \n",
       "5         -0.1610    -0.1200    -0.1520   -0.02830  ...     0.0991    -0.0738   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "243500    -0.2440     0.0160    -0.1790   -0.34600  ...     0.1480    -0.1520   \n",
       "243501     0.0000     0.0000     0.0433    0.00000  ...     0.2160     0.0000   \n",
       "243502     0.0000     0.0000     0.0000    0.00000  ...     0.0000    -0.2540   \n",
       "243503    -0.0938    -0.0330    -0.1030    0.00613  ...     0.0709     0.0626   \n",
       "243504    -0.0797    -0.0113    -0.1080   -0.00240  ...     0.0639     0.0483   \n",
       "\n",
       "        GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  GSM540608  \\\n",
       "ID                                                                         \n",
       "1        0.002070     0.0208    0.00841     0.0122   -0.01280     0.0748   \n",
       "2        0.000000     0.0000    0.00000     0.0000    0.00000    -0.5670   \n",
       "3        0.000000     0.0000    0.00000     0.0000    0.00000    -0.4420   \n",
       "4        0.364000     0.0321   -0.24800     0.0240   -0.08650     0.0336   \n",
       "5       -0.058400     0.0120   -0.06350    -0.1470   -0.24700     0.1170   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "243500  -0.211000    -0.0559    0.08870    -0.0608    0.00423    -0.4000   \n",
       "243501   0.122000     0.0000    0.00000     0.0000    0.00000     0.0000   \n",
       "243502   0.000000     0.0000    0.00000     0.0000    0.00000    -0.1480   \n",
       "243503  -0.000328     0.0182    0.01140     0.0342   -0.03030     0.0543   \n",
       "243504  -0.002440     0.0539    0.02960     0.0264   -0.00543     0.0700   \n",
       "\n",
       "        GSM540609  GSM540610  \n",
       "ID                            \n",
       "1       -0.000229    -0.0148  \n",
       "2       -0.004620    -0.4700  \n",
       "3        0.000000    -0.6850  \n",
       "4        0.167000     0.0337  \n",
       "5       -0.115000     0.0135  \n",
       "...           ...        ...  \n",
       "243500  -0.069500    -0.0305  \n",
       "243501   0.000000    -1.0400  \n",
       "243502  -0.302000    -0.9700  \n",
       "243503   0.014000    -0.0138  \n",
       "243504  -0.012400    -0.0203  \n",
       "\n",
       "[243429 rows x 61 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1', '2', '3', '4', '5'], 'COL': ['267', '267', '267', '267', '267'], 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0], 'SPOT_ID': ['HsCGHBrightCorner', 'DarkCorner', 'DarkCorner', 'A_16_P20527812', 'A_16_P01708709'], 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'], 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'], 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'], 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'], 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'], 'CHROMOSOMAL_LOCATION': [nan, nan, nan, 'chr16:076331867-076331926', 'chr7:047626734-047626793'], 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'], 'DESCRIPTION': [nan, nan, nan, nan, 'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'], 'GB_RANGE': [nan, nan, nan, 'NC_000016.8[076331867..076331926]', 'NC_000007.11[047626734..047626793]']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '13', '14', '15', '16', '17', '18', '19', '20', '21']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1', '2', '3', '4', '5'], 'COL': ['267', '267', '267', '267', '267'], 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0], 'SPOT_ID': ['HsCGHBrightCorner', 'DarkCorner', 'DarkCorner', 'A_16_P20527812', 'A_16_P01708709'], 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'], 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'], 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'], 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'], 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'], 'CHROMOSOMAL_LOCATION': [nan, nan, nan, 'chr16:076331867-076331926', 'chr7:047626734-047626793'], 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'], 'DESCRIPTION': [nan, nan, nan, nan, 'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'], 'GB_RANGE': [nan, nan, nan, 'NC_000016.8[076331867..076331926]', 'NC_000007.11[047626734..047626793]']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>COL</th>\n",
       "      <th>ROW</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>CONTROL_TYPE</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ACCESSION_STRING</th>\n",
       "      <th>CHROMOSOMAL_LOCATION</th>\n",
       "      <th>CYTOBAND</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>GB_RANGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>267</td>\n",
       "      <td>912.0</td>\n",
       "      <td>HsCGHBrightCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>267</td>\n",
       "      <td>910.0</td>\n",
       "      <td>DarkCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>267</td>\n",
       "      <td>908.0</td>\n",
       "      <td>DarkCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>267</td>\n",
       "      <td>906.0</td>\n",
       "      <td>A_16_P20527812</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16:076331867-076331926</td>\n",
       "      <td>hs|q23.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000016.8[076331867..076331926]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>267</td>\n",
       "      <td>904.0</td>\n",
       "      <td>A_16_P01708709</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NM_138295</td>\n",
       "      <td>PKD1L1</td>\n",
       "      <td>polycystic kidney disease 1 like 1</td>\n",
       "      <td>ref|NM_138295|ref|NM_025031</td>\n",
       "      <td>chr7:047626734-047626793</td>\n",
       "      <td>hs|p12.3</td>\n",
       "      <td>Homo sapiens polycystic kidney disease 1 like ...</td>\n",
       "      <td>NC_000007.11[047626734..047626793]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092790</th>\n",
       "      <td>243500</td>\n",
       "      <td>-3.05E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092791</th>\n",
       "      <td>243501</td>\n",
       "      <td>-1.04E+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092792</th>\n",
       "      <td>243502</td>\n",
       "      <td>-9.70E-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092793</th>\n",
       "      <td>243503</td>\n",
       "      <td>-1.38E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092794</th>\n",
       "      <td>243504</td>\n",
       "      <td>-2.03E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15092795 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID        COL    ROW            SPOT_ID CONTROL_TYPE     GB_ACC  \\\n",
       "0              1        267  912.0  HsCGHBrightCorner          pos        NaN   \n",
       "1              2        267  910.0         DarkCorner          pos        NaN   \n",
       "2              3        267  908.0         DarkCorner          pos        NaN   \n",
       "3              4        267  906.0     A_16_P20527812        FALSE        NaN   \n",
       "4              5        267  904.0     A_16_P01708709        FALSE  NM_138295   \n",
       "...          ...        ...    ...                ...          ...        ...   \n",
       "15092790  243500  -3.05E-02    NaN                NaN          NaN        NaN   \n",
       "15092791  243501  -1.04E+00    NaN                NaN          NaN        NaN   \n",
       "15092792  243502  -9.70E-01    NaN                NaN          NaN        NaN   \n",
       "15092793  243503  -1.38E-02    NaN                NaN          NaN        NaN   \n",
       "15092794  243504  -2.03E-02    NaN                NaN          NaN        NaN   \n",
       "\n",
       "         GENE_SYMBOL                           GENE_NAME  \\\n",
       "0                NaN                                 NaN   \n",
       "1                NaN                                 NaN   \n",
       "2                NaN                                 NaN   \n",
       "3                NaN                                 NaN   \n",
       "4             PKD1L1  polycystic kidney disease 1 like 1   \n",
       "...              ...                                 ...   \n",
       "15092790         NaN                                 NaN   \n",
       "15092791         NaN                                 NaN   \n",
       "15092792         NaN                                 NaN   \n",
       "15092793         NaN                                 NaN   \n",
       "15092794         NaN                                 NaN   \n",
       "\n",
       "                     ACCESSION_STRING       CHROMOSOMAL_LOCATION  CYTOBAND  \\\n",
       "0                                 NaN                        NaN       NaN   \n",
       "1                                 NaN                        NaN       NaN   \n",
       "2                                 NaN                        NaN       NaN   \n",
       "3                                 NaN  chr16:076331867-076331926  hs|q23.1   \n",
       "4         ref|NM_138295|ref|NM_025031   chr7:047626734-047626793  hs|p12.3   \n",
       "...                               ...                        ...       ...   \n",
       "15092790                          NaN                        NaN       NaN   \n",
       "15092791                          NaN                        NaN       NaN   \n",
       "15092792                          NaN                        NaN       NaN   \n",
       "15092793                          NaN                        NaN       NaN   \n",
       "15092794                          NaN                        NaN       NaN   \n",
       "\n",
       "                                                DESCRIPTION  \\\n",
       "0                                                       NaN   \n",
       "1                                                       NaN   \n",
       "2                                                       NaN   \n",
       "3                                                       NaN   \n",
       "4         Homo sapiens polycystic kidney disease 1 like ...   \n",
       "...                                                     ...   \n",
       "15092790                                                NaN   \n",
       "15092791                                                NaN   \n",
       "15092792                                                NaN   \n",
       "15092793                                                NaN   \n",
       "15092794                                                NaN   \n",
       "\n",
       "                                    GB_RANGE  \n",
       "0                                        NaN  \n",
       "1                                        NaN  \n",
       "2                                        NaN  \n",
       "3          NC_000016.8[076331867..076331926]  \n",
       "4         NC_000007.11[047626734..047626793]  \n",
       "...                                      ...  \n",
       "15092790                                 NaN  \n",
       "15092791                                 NaN  \n",
       "15092792                                 NaN  \n",
       "15092793                                 NaN  \n",
       "15092794                                 NaN  \n",
       "\n",
       "[15092795 rows x 13 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'COL', 'ROW', 'SPOT_ID', 'CONTROL_TYPE', 'GB_ACC', 'GENE_SYMBOL',\n",
       "       'GENE_NAME', 'ACCESSION_STRING', 'CHROMOSOMAL_LOCATION', 'CYTOBAND',\n",
       "       'DESCRIPTION', 'GB_RANGE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GENE_SYMBOL'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>-0.091366</td>\n",
       "      <td>-0.009954</td>\n",
       "      <td>0.064026</td>\n",
       "      <td>0.223229</td>\n",
       "      <td>0.015514</td>\n",
       "      <td>0.133943</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>0.036996</td>\n",
       "      <td>-0.109300</td>\n",
       "      <td>0.058823</td>\n",
       "      <td>...</td>\n",
       "      <td>0.152000</td>\n",
       "      <td>0.040586</td>\n",
       "      <td>0.114029</td>\n",
       "      <td>-0.021514</td>\n",
       "      <td>0.110771</td>\n",
       "      <td>0.135557</td>\n",
       "      <td>0.125829</td>\n",
       "      <td>0.009514</td>\n",
       "      <td>0.174671</td>\n",
       "      <td>0.049243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15E1.2</th>\n",
       "      <td>-0.018767</td>\n",
       "      <td>-0.036233</td>\n",
       "      <td>0.126533</td>\n",
       "      <td>0.015967</td>\n",
       "      <td>-0.001400</td>\n",
       "      <td>-0.229000</td>\n",
       "      <td>0.095173</td>\n",
       "      <td>-0.137293</td>\n",
       "      <td>-0.145300</td>\n",
       "      <td>0.137067</td>\n",
       "      <td>...</td>\n",
       "      <td>0.146700</td>\n",
       "      <td>0.267000</td>\n",
       "      <td>0.207333</td>\n",
       "      <td>0.039400</td>\n",
       "      <td>-0.047812</td>\n",
       "      <td>0.011360</td>\n",
       "      <td>-0.109767</td>\n",
       "      <td>0.300333</td>\n",
       "      <td>0.013167</td>\n",
       "      <td>-0.008633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2'-PDE</th>\n",
       "      <td>-0.095650</td>\n",
       "      <td>-0.038200</td>\n",
       "      <td>-0.273500</td>\n",
       "      <td>-0.188500</td>\n",
       "      <td>0.077200</td>\n",
       "      <td>-0.170500</td>\n",
       "      <td>-0.023250</td>\n",
       "      <td>-0.265000</td>\n",
       "      <td>-0.224050</td>\n",
       "      <td>0.028700</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.189500</td>\n",
       "      <td>-0.117450</td>\n",
       "      <td>-0.166800</td>\n",
       "      <td>-0.037500</td>\n",
       "      <td>-0.123500</td>\n",
       "      <td>0.019000</td>\n",
       "      <td>-0.111050</td>\n",
       "      <td>0.182000</td>\n",
       "      <td>-0.032200</td>\n",
       "      <td>-0.052400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76P</th>\n",
       "      <td>-0.177225</td>\n",
       "      <td>-0.016975</td>\n",
       "      <td>-0.015500</td>\n",
       "      <td>-0.230250</td>\n",
       "      <td>-0.381500</td>\n",
       "      <td>-0.132625</td>\n",
       "      <td>-0.074625</td>\n",
       "      <td>0.013000</td>\n",
       "      <td>-0.158050</td>\n",
       "      <td>-0.093265</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067575</td>\n",
       "      <td>0.068450</td>\n",
       "      <td>-0.156500</td>\n",
       "      <td>0.048800</td>\n",
       "      <td>-0.054375</td>\n",
       "      <td>-0.010340</td>\n",
       "      <td>-0.110975</td>\n",
       "      <td>0.131200</td>\n",
       "      <td>-0.107950</td>\n",
       "      <td>-0.035025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7A5</th>\n",
       "      <td>-0.060111</td>\n",
       "      <td>0.072104</td>\n",
       "      <td>0.055131</td>\n",
       "      <td>-0.020300</td>\n",
       "      <td>0.201556</td>\n",
       "      <td>0.109311</td>\n",
       "      <td>0.029308</td>\n",
       "      <td>0.029044</td>\n",
       "      <td>0.107101</td>\n",
       "      <td>-0.007541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.151611</td>\n",
       "      <td>0.148789</td>\n",
       "      <td>0.090900</td>\n",
       "      <td>-0.000616</td>\n",
       "      <td>0.151533</td>\n",
       "      <td>0.079336</td>\n",
       "      <td>-0.004712</td>\n",
       "      <td>-0.077400</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>0.031200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bA9F11.1</th>\n",
       "      <td>0.150150</td>\n",
       "      <td>0.095300</td>\n",
       "      <td>-0.018000</td>\n",
       "      <td>0.263000</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>-0.021791</td>\n",
       "      <td>0.147900</td>\n",
       "      <td>-0.086950</td>\n",
       "      <td>-0.105000</td>\n",
       "      <td>-0.064150</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.048850</td>\n",
       "      <td>0.065850</td>\n",
       "      <td>0.111600</td>\n",
       "      <td>-0.127850</td>\n",
       "      <td>-0.217500</td>\n",
       "      <td>-0.151800</td>\n",
       "      <td>-0.283000</td>\n",
       "      <td>-0.072160</td>\n",
       "      <td>0.191000</td>\n",
       "      <td>0.003150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <td>0.006610</td>\n",
       "      <td>-0.161000</td>\n",
       "      <td>-0.120000</td>\n",
       "      <td>-0.133000</td>\n",
       "      <td>-0.138000</td>\n",
       "      <td>-0.177000</td>\n",
       "      <td>-0.132000</td>\n",
       "      <td>0.051200</td>\n",
       "      <td>-0.039200</td>\n",
       "      <td>-0.014300</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.188000</td>\n",
       "      <td>-0.012400</td>\n",
       "      <td>0.119000</td>\n",
       "      <td>-0.193000</td>\n",
       "      <td>-0.092300</td>\n",
       "      <td>-0.124000</td>\n",
       "      <td>-0.271000</td>\n",
       "      <td>-0.176000</td>\n",
       "      <td>-0.100000</td>\n",
       "      <td>-0.269000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-D3</th>\n",
       "      <td>-0.104746</td>\n",
       "      <td>-0.010522</td>\n",
       "      <td>0.026106</td>\n",
       "      <td>0.072607</td>\n",
       "      <td>-0.014310</td>\n",
       "      <td>0.027170</td>\n",
       "      <td>0.043281</td>\n",
       "      <td>0.107378</td>\n",
       "      <td>0.029616</td>\n",
       "      <td>-0.188333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.173267</td>\n",
       "      <td>-0.299222</td>\n",
       "      <td>-0.090567</td>\n",
       "      <td>-0.004926</td>\n",
       "      <td>0.055044</td>\n",
       "      <td>0.054733</td>\n",
       "      <td>0.012011</td>\n",
       "      <td>0.007011</td>\n",
       "      <td>0.017700</td>\n",
       "      <td>0.026992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hCAP-H2</th>\n",
       "      <td>0.053800</td>\n",
       "      <td>0.027662</td>\n",
       "      <td>0.011612</td>\n",
       "      <td>0.063725</td>\n",
       "      <td>-0.114200</td>\n",
       "      <td>0.026400</td>\n",
       "      <td>0.140775</td>\n",
       "      <td>-0.134775</td>\n",
       "      <td>-0.026850</td>\n",
       "      <td>0.035800</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.031643</td>\n",
       "      <td>-0.244000</td>\n",
       "      <td>0.083365</td>\n",
       "      <td>-0.079825</td>\n",
       "      <td>-0.076025</td>\n",
       "      <td>-0.071883</td>\n",
       "      <td>-0.031462</td>\n",
       "      <td>0.068550</td>\n",
       "      <td>0.098075</td>\n",
       "      <td>0.060400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mimitin</th>\n",
       "      <td>-0.052605</td>\n",
       "      <td>0.082836</td>\n",
       "      <td>0.032056</td>\n",
       "      <td>-0.053923</td>\n",
       "      <td>-0.005092</td>\n",
       "      <td>-0.007739</td>\n",
       "      <td>-0.010476</td>\n",
       "      <td>0.012062</td>\n",
       "      <td>-0.002436</td>\n",
       "      <td>-0.199143</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130712</td>\n",
       "      <td>0.100208</td>\n",
       "      <td>0.085496</td>\n",
       "      <td>0.004645</td>\n",
       "      <td>0.171510</td>\n",
       "      <td>-0.019704</td>\n",
       "      <td>0.012892</td>\n",
       "      <td>-0.295004</td>\n",
       "      <td>0.053736</td>\n",
       "      <td>-0.030301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16832 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "Gene                                                                           \n",
       "            -0.091366  -0.009954   0.064026   0.223229   0.015514   0.133943   \n",
       "15E1.2      -0.018767  -0.036233   0.126533   0.015967  -0.001400  -0.229000   \n",
       "2'-PDE      -0.095650  -0.038200  -0.273500  -0.188500   0.077200  -0.170500   \n",
       "76P         -0.177225  -0.016975  -0.015500  -0.230250  -0.381500  -0.132625   \n",
       "7A5         -0.060111   0.072104   0.055131  -0.020300   0.201556   0.109311   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "bA9F11.1     0.150150   0.095300  -0.018000   0.263000   0.034750  -0.021791   \n",
       "dJ341D10.1   0.006610  -0.161000  -0.120000  -0.133000  -0.138000  -0.177000   \n",
       "hCAP-D3     -0.104746  -0.010522   0.026106   0.072607  -0.014310   0.027170   \n",
       "hCAP-H2      0.053800   0.027662   0.011612   0.063725  -0.114200   0.026400   \n",
       "mimitin     -0.052605   0.082836   0.032056  -0.053923  -0.005092  -0.007739   \n",
       "\n",
       "            GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  \\\n",
       "Gene                                                    ...              \n",
       "             0.010000   0.036996  -0.109300   0.058823  ...   0.152000   \n",
       "15E1.2       0.095173  -0.137293  -0.145300   0.137067  ...   0.146700   \n",
       "2'-PDE      -0.023250  -0.265000  -0.224050   0.028700  ...  -0.189500   \n",
       "76P         -0.074625   0.013000  -0.158050  -0.093265  ...   0.067575   \n",
       "7A5          0.029308   0.029044   0.107101  -0.007541  ...   0.151611   \n",
       "...               ...        ...        ...        ...  ...        ...   \n",
       "bA9F11.1     0.147900  -0.086950  -0.105000  -0.064150  ...  -0.048850   \n",
       "dJ341D10.1  -0.132000   0.051200  -0.039200  -0.014300  ...  -0.188000   \n",
       "hCAP-D3      0.043281   0.107378   0.029616  -0.188333  ...   0.173267   \n",
       "hCAP-H2      0.140775  -0.134775  -0.026850   0.035800  ...  -0.031643   \n",
       "mimitin     -0.010476   0.012062  -0.002436  -0.199143  ...   0.130712   \n",
       "\n",
       "            GSM540602  GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  \\\n",
       "Gene                                                                           \n",
       "             0.040586   0.114029  -0.021514   0.110771   0.135557   0.125829   \n",
       "15E1.2       0.267000   0.207333   0.039400  -0.047812   0.011360  -0.109767   \n",
       "2'-PDE      -0.117450  -0.166800  -0.037500  -0.123500   0.019000  -0.111050   \n",
       "76P          0.068450  -0.156500   0.048800  -0.054375  -0.010340  -0.110975   \n",
       "7A5          0.148789   0.090900  -0.000616   0.151533   0.079336  -0.004712   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "bA9F11.1     0.065850   0.111600  -0.127850  -0.217500  -0.151800  -0.283000   \n",
       "dJ341D10.1  -0.012400   0.119000  -0.193000  -0.092300  -0.124000  -0.271000   \n",
       "hCAP-D3     -0.299222  -0.090567  -0.004926   0.055044   0.054733   0.012011   \n",
       "hCAP-H2     -0.244000   0.083365  -0.079825  -0.076025  -0.071883  -0.031462   \n",
       "mimitin      0.100208   0.085496   0.004645   0.171510  -0.019704   0.012892   \n",
       "\n",
       "            GSM540608  GSM540609  GSM540610  \n",
       "Gene                                         \n",
       "             0.009514   0.174671   0.049243  \n",
       "15E1.2       0.300333   0.013167  -0.008633  \n",
       "2'-PDE       0.182000  -0.032200  -0.052400  \n",
       "76P          0.131200  -0.107950  -0.035025  \n",
       "7A5         -0.077400  -0.000033   0.031200  \n",
       "...               ...        ...        ...  \n",
       "bA9F11.1    -0.072160   0.191000   0.003150  \n",
       "dJ341D10.1  -0.176000  -0.100000  -0.269000  \n",
       "hCAP-D3      0.007011   0.017700   0.026992  \n",
       "hCAP-H2      0.068550   0.098075   0.060400  \n",
       "mimitin     -0.295004   0.053736  -0.030301  \n",
       "\n",
       "[16832 rows x 61 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.261000</td>\n",
       "      <td>0.023700</td>\n",
       "      <td>0.117000</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>0.141000</td>\n",
       "      <td>0.239000</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>0.216000</td>\n",
       "      <td>0.318000</td>\n",
       "      <td>0.085300</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.164000</td>\n",
       "      <td>0.109000</td>\n",
       "      <td>0.205000</td>\n",
       "      <td>-0.053900</td>\n",
       "      <td>-0.069000</td>\n",
       "      <td>-0.034000</td>\n",
       "      <td>0.054000</td>\n",
       "      <td>-0.154000</td>\n",
       "      <td>0.242000</td>\n",
       "      <td>0.074100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-0.071700</td>\n",
       "      <td>-0.075717</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>0.176317</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>0.049178</td>\n",
       "      <td>-0.039567</td>\n",
       "      <td>0.045017</td>\n",
       "      <td>0.107967</td>\n",
       "      <td>0.116025</td>\n",
       "      <td>...</td>\n",
       "      <td>0.539667</td>\n",
       "      <td>0.245167</td>\n",
       "      <td>0.143050</td>\n",
       "      <td>0.039052</td>\n",
       "      <td>-0.059468</td>\n",
       "      <td>0.115183</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>0.168233</td>\n",
       "      <td>0.139317</td>\n",
       "      <td>0.074317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.056388</td>\n",
       "      <td>-0.025460</td>\n",
       "      <td>0.121440</td>\n",
       "      <td>0.166813</td>\n",
       "      <td>0.002068</td>\n",
       "      <td>0.105455</td>\n",
       "      <td>0.087587</td>\n",
       "      <td>0.114687</td>\n",
       "      <td>0.150363</td>\n",
       "      <td>0.049269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.537412</td>\n",
       "      <td>0.228625</td>\n",
       "      <td>0.226325</td>\n",
       "      <td>0.063661</td>\n",
       "      <td>0.065370</td>\n",
       "      <td>-0.013030</td>\n",
       "      <td>-0.029762</td>\n",
       "      <td>0.093887</td>\n",
       "      <td>0.101675</td>\n",
       "      <td>-0.023225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.115900</td>\n",
       "      <td>0.056557</td>\n",
       "      <td>-0.057271</td>\n",
       "      <td>0.139629</td>\n",
       "      <td>-0.060130</td>\n",
       "      <td>-0.018657</td>\n",
       "      <td>0.206329</td>\n",
       "      <td>-0.027757</td>\n",
       "      <td>-0.015286</td>\n",
       "      <td>-0.037541</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.090871</td>\n",
       "      <td>-0.217963</td>\n",
       "      <td>-0.032471</td>\n",
       "      <td>-0.165701</td>\n",
       "      <td>-0.124071</td>\n",
       "      <td>-0.120486</td>\n",
       "      <td>-0.127914</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>0.150286</td>\n",
       "      <td>-0.038199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>0.185500</td>\n",
       "      <td>-0.096250</td>\n",
       "      <td>-0.289000</td>\n",
       "      <td>-0.030060</td>\n",
       "      <td>-0.063700</td>\n",
       "      <td>-0.112750</td>\n",
       "      <td>-0.056290</td>\n",
       "      <td>0.054900</td>\n",
       "      <td>0.011600</td>\n",
       "      <td>0.062265</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.253000</td>\n",
       "      <td>0.114100</td>\n",
       "      <td>-0.105850</td>\n",
       "      <td>-0.118550</td>\n",
       "      <td>-0.061350</td>\n",
       "      <td>0.013800</td>\n",
       "      <td>0.006250</td>\n",
       "      <td>-0.114600</td>\n",
       "      <td>-0.029010</td>\n",
       "      <td>-0.064800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>0.021000</td>\n",
       "      <td>-0.025660</td>\n",
       "      <td>-0.214200</td>\n",
       "      <td>-0.152360</td>\n",
       "      <td>-0.020380</td>\n",
       "      <td>-0.004088</td>\n",
       "      <td>-0.022060</td>\n",
       "      <td>0.012460</td>\n",
       "      <td>0.035980</td>\n",
       "      <td>-0.095848</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.150060</td>\n",
       "      <td>-0.114840</td>\n",
       "      <td>-0.113584</td>\n",
       "      <td>-0.069040</td>\n",
       "      <td>-0.112720</td>\n",
       "      <td>-0.080224</td>\n",
       "      <td>-0.191920</td>\n",
       "      <td>-0.193660</td>\n",
       "      <td>-0.168400</td>\n",
       "      <td>-0.114120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.018978</td>\n",
       "      <td>-0.048632</td>\n",
       "      <td>-0.240545</td>\n",
       "      <td>-0.211436</td>\n",
       "      <td>-0.014087</td>\n",
       "      <td>-0.029066</td>\n",
       "      <td>0.012171</td>\n",
       "      <td>0.010093</td>\n",
       "      <td>0.011982</td>\n",
       "      <td>-0.161527</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.117764</td>\n",
       "      <td>-0.096468</td>\n",
       "      <td>-0.157382</td>\n",
       "      <td>-0.115042</td>\n",
       "      <td>-0.050454</td>\n",
       "      <td>-0.110373</td>\n",
       "      <td>-0.230038</td>\n",
       "      <td>-0.178318</td>\n",
       "      <td>-0.209545</td>\n",
       "      <td>-0.195591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>-0.034700</td>\n",
       "      <td>0.021967</td>\n",
       "      <td>0.089867</td>\n",
       "      <td>0.084667</td>\n",
       "      <td>-0.098300</td>\n",
       "      <td>0.109900</td>\n",
       "      <td>0.071067</td>\n",
       "      <td>-0.106867</td>\n",
       "      <td>-0.011533</td>\n",
       "      <td>0.158667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.127000</td>\n",
       "      <td>0.010613</td>\n",
       "      <td>0.163800</td>\n",
       "      <td>-0.020167</td>\n",
       "      <td>-0.006195</td>\n",
       "      <td>0.057200</td>\n",
       "      <td>0.041467</td>\n",
       "      <td>0.072733</td>\n",
       "      <td>0.206333</td>\n",
       "      <td>0.196667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>-0.129433</td>\n",
       "      <td>-0.158727</td>\n",
       "      <td>-0.175639</td>\n",
       "      <td>-0.123181</td>\n",
       "      <td>-0.021133</td>\n",
       "      <td>-0.288200</td>\n",
       "      <td>-0.128013</td>\n",
       "      <td>-0.138720</td>\n",
       "      <td>-0.097573</td>\n",
       "      <td>0.044883</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.167607</td>\n",
       "      <td>-0.057523</td>\n",
       "      <td>-0.091887</td>\n",
       "      <td>0.025984</td>\n",
       "      <td>-0.078663</td>\n",
       "      <td>-0.026909</td>\n",
       "      <td>-0.100020</td>\n",
       "      <td>0.083072</td>\n",
       "      <td>-0.234480</td>\n",
       "      <td>-0.077160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>-0.079166</td>\n",
       "      <td>-0.043763</td>\n",
       "      <td>-0.262375</td>\n",
       "      <td>-0.159617</td>\n",
       "      <td>0.018503</td>\n",
       "      <td>-0.025875</td>\n",
       "      <td>-0.000616</td>\n",
       "      <td>0.043308</td>\n",
       "      <td>0.033308</td>\n",
       "      <td>-0.232042</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.136790</td>\n",
       "      <td>-0.071925</td>\n",
       "      <td>-0.230033</td>\n",
       "      <td>-0.103192</td>\n",
       "      <td>0.032115</td>\n",
       "      <td>-0.087575</td>\n",
       "      <td>-0.126593</td>\n",
       "      <td>-0.300383</td>\n",
       "      <td>-0.298500</td>\n",
       "      <td>-0.095075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12669 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "A1BG     0.261000   0.023700   0.117000   0.154000   0.141000   0.239000   \n",
       "A2M     -0.071700  -0.075717   0.184000   0.176317   0.034750   0.049178   \n",
       "A2ML1   -0.056388  -0.025460   0.121440   0.166813   0.002068   0.105455   \n",
       "A4GALT   0.115900   0.056557  -0.057271   0.139629  -0.060130  -0.018657   \n",
       "A4GNT    0.185500  -0.096250  -0.289000  -0.030060  -0.063700  -0.112750   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A   0.021000  -0.025660  -0.214200  -0.152360  -0.020380  -0.004088   \n",
       "ZYG11B   0.018978  -0.048632  -0.240545  -0.211436  -0.014087  -0.029066   \n",
       "ZYX     -0.034700   0.021967   0.089867   0.084667  -0.098300   0.109900   \n",
       "ZZEF1   -0.129433  -0.158727  -0.175639  -0.123181  -0.021133  -0.288200   \n",
       "ZZZ3    -0.079166  -0.043763  -0.262375  -0.159617   0.018503  -0.025875   \n",
       "\n",
       "        GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  GSM540602  \\\n",
       "A1BG     0.197000   0.216000   0.318000   0.085300  ...  -0.164000   0.109000   \n",
       "A2M     -0.039567   0.045017   0.107967   0.116025  ...   0.539667   0.245167   \n",
       "A2ML1    0.087587   0.114687   0.150363   0.049269  ...   0.537412   0.228625   \n",
       "A4GALT   0.206329  -0.027757  -0.015286  -0.037541  ...  -0.090871  -0.217963   \n",
       "A4GNT   -0.056290   0.054900   0.011600   0.062265  ...  -0.253000   0.114100   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "ZYG11A  -0.022060   0.012460   0.035980  -0.095848  ...  -0.150060  -0.114840   \n",
       "ZYG11B   0.012171   0.010093   0.011982  -0.161527  ...  -0.117764  -0.096468   \n",
       "ZYX      0.071067  -0.106867  -0.011533   0.158667  ...   0.127000   0.010613   \n",
       "ZZEF1   -0.128013  -0.138720  -0.097573   0.044883  ...  -0.167607  -0.057523   \n",
       "ZZZ3    -0.000616   0.043308   0.033308  -0.232042  ...  -0.136790  -0.071925   \n",
       "\n",
       "        GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  GSM540608  \\\n",
       "A1BG     0.205000  -0.053900  -0.069000  -0.034000   0.054000  -0.154000   \n",
       "A2M      0.143050   0.039052  -0.059468   0.115183   0.112083   0.168233   \n",
       "A2ML1    0.226325   0.063661   0.065370  -0.013030  -0.029762   0.093887   \n",
       "A4GALT  -0.032471  -0.165701  -0.124071  -0.120486  -0.127914   0.012549   \n",
       "A4GNT   -0.105850  -0.118550  -0.061350   0.013800   0.006250  -0.114600   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A  -0.113584  -0.069040  -0.112720  -0.080224  -0.191920  -0.193660   \n",
       "ZYG11B  -0.157382  -0.115042  -0.050454  -0.110373  -0.230038  -0.178318   \n",
       "ZYX      0.163800  -0.020167  -0.006195   0.057200   0.041467   0.072733   \n",
       "ZZEF1   -0.091887   0.025984  -0.078663  -0.026909  -0.100020   0.083072   \n",
       "ZZZ3    -0.230033  -0.103192   0.032115  -0.087575  -0.126593  -0.300383   \n",
       "\n",
       "        GSM540609  GSM540610  \n",
       "A1BG     0.242000   0.074100  \n",
       "A2M      0.139317   0.074317  \n",
       "A2ML1    0.101675  -0.023225  \n",
       "A4GALT   0.150286  -0.038199  \n",
       "A4GNT   -0.029010  -0.064800  \n",
       "...           ...        ...  \n",
       "ZYG11A  -0.168400  -0.114120  \n",
       "ZYG11B  -0.209545  -0.195591  \n",
       "ZYX      0.206333   0.196667  \n",
       "ZZEF1   -0.234480  -0.077160  \n",
       "ZZZ3    -0.298500  -0.095075  \n",
       "\n",
       "[12669 rows x 61 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM540550</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.071700</td>\n",
       "      <td>-0.056388</td>\n",
       "      <td>0.115900</td>\n",
       "      <td>0.18550</td>\n",
       "      <td>-0.074332</td>\n",
       "      <td>0.284667</td>\n",
       "      <td>-0.050671</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011806</td>\n",
       "      <td>-0.069583</td>\n",
       "      <td>0.0818</td>\n",
       "      <td>-0.26000</td>\n",
       "      <td>0.19156</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>0.018978</td>\n",
       "      <td>-0.034700</td>\n",
       "      <td>-0.129433</td>\n",
       "      <td>-0.079166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540551</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0237</td>\n",
       "      <td>-0.075717</td>\n",
       "      <td>-0.025460</td>\n",
       "      <td>0.056557</td>\n",
       "      <td>-0.09625</td>\n",
       "      <td>0.058537</td>\n",
       "      <td>0.048833</td>\n",
       "      <td>-0.124633</td>\n",
       "      <td>-0.00876</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074980</td>\n",
       "      <td>0.105617</td>\n",
       "      <td>0.1890</td>\n",
       "      <td>-0.39000</td>\n",
       "      <td>-0.02402</td>\n",
       "      <td>-0.025660</td>\n",
       "      <td>-0.048632</td>\n",
       "      <td>0.021967</td>\n",
       "      <td>-0.158727</td>\n",
       "      <td>-0.043763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540552</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>0.121440</td>\n",
       "      <td>-0.057271</td>\n",
       "      <td>-0.28900</td>\n",
       "      <td>0.049406</td>\n",
       "      <td>0.168933</td>\n",
       "      <td>0.118744</td>\n",
       "      <td>-0.20600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117400</td>\n",
       "      <td>0.014768</td>\n",
       "      <td>0.1560</td>\n",
       "      <td>-0.51600</td>\n",
       "      <td>-0.26312</td>\n",
       "      <td>-0.214200</td>\n",
       "      <td>-0.240545</td>\n",
       "      <td>0.089867</td>\n",
       "      <td>-0.175639</td>\n",
       "      <td>-0.262375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540553</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.1540</td>\n",
       "      <td>0.176317</td>\n",
       "      <td>0.166813</td>\n",
       "      <td>0.139629</td>\n",
       "      <td>-0.03006</td>\n",
       "      <td>-0.036533</td>\n",
       "      <td>0.314667</td>\n",
       "      <td>0.071756</td>\n",
       "      <td>0.12700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.120940</td>\n",
       "      <td>0.090387</td>\n",
       "      <td>0.0823</td>\n",
       "      <td>-0.42100</td>\n",
       "      <td>0.12542</td>\n",
       "      <td>-0.152360</td>\n",
       "      <td>-0.211436</td>\n",
       "      <td>0.084667</td>\n",
       "      <td>-0.123181</td>\n",
       "      <td>-0.159617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540554</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.1410</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>0.002068</td>\n",
       "      <td>-0.060130</td>\n",
       "      <td>-0.06370</td>\n",
       "      <td>0.322153</td>\n",
       "      <td>0.049267</td>\n",
       "      <td>-0.056361</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015800</td>\n",
       "      <td>0.022417</td>\n",
       "      <td>0.0821</td>\n",
       "      <td>-0.58000</td>\n",
       "      <td>0.07364</td>\n",
       "      <td>-0.020380</td>\n",
       "      <td>-0.014087</td>\n",
       "      <td>-0.098300</td>\n",
       "      <td>-0.021133</td>\n",
       "      <td>0.018503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540606</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.0340</td>\n",
       "      <td>0.115183</td>\n",
       "      <td>-0.013030</td>\n",
       "      <td>-0.120486</td>\n",
       "      <td>0.01380</td>\n",
       "      <td>0.096020</td>\n",
       "      <td>-0.082280</td>\n",
       "      <td>-0.084410</td>\n",
       "      <td>0.14000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053585</td>\n",
       "      <td>0.004537</td>\n",
       "      <td>0.0128</td>\n",
       "      <td>0.00522</td>\n",
       "      <td>-0.05314</td>\n",
       "      <td>-0.080224</td>\n",
       "      <td>-0.110373</td>\n",
       "      <td>0.057200</td>\n",
       "      <td>-0.026909</td>\n",
       "      <td>-0.087575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540607</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>-0.029762</td>\n",
       "      <td>-0.127914</td>\n",
       "      <td>0.00625</td>\n",
       "      <td>0.126520</td>\n",
       "      <td>-0.074100</td>\n",
       "      <td>-0.090500</td>\n",
       "      <td>0.15000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000366</td>\n",
       "      <td>-0.128767</td>\n",
       "      <td>0.0391</td>\n",
       "      <td>0.53800</td>\n",
       "      <td>-0.00796</td>\n",
       "      <td>-0.191920</td>\n",
       "      <td>-0.230038</td>\n",
       "      <td>0.041467</td>\n",
       "      <td>-0.100020</td>\n",
       "      <td>-0.126593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540608</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.1540</td>\n",
       "      <td>0.168233</td>\n",
       "      <td>0.093887</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>-0.11460</td>\n",
       "      <td>-0.061250</td>\n",
       "      <td>0.163033</td>\n",
       "      <td>0.201889</td>\n",
       "      <td>0.07380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.075702</td>\n",
       "      <td>0.154717</td>\n",
       "      <td>0.1470</td>\n",
       "      <td>-0.50600</td>\n",
       "      <td>0.14348</td>\n",
       "      <td>-0.193660</td>\n",
       "      <td>-0.178318</td>\n",
       "      <td>0.072733</td>\n",
       "      <td>0.083072</td>\n",
       "      <td>-0.300383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540609</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2420</td>\n",
       "      <td>0.139317</td>\n",
       "      <td>0.101675</td>\n",
       "      <td>0.150286</td>\n",
       "      <td>-0.02901</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>0.065833</td>\n",
       "      <td>-0.030744</td>\n",
       "      <td>0.07190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.066480</td>\n",
       "      <td>-0.007978</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>-0.48600</td>\n",
       "      <td>0.05238</td>\n",
       "      <td>-0.168400</td>\n",
       "      <td>-0.209545</td>\n",
       "      <td>0.206333</td>\n",
       "      <td>-0.234480</td>\n",
       "      <td>-0.298500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540610</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0741</td>\n",
       "      <td>0.074317</td>\n",
       "      <td>-0.023225</td>\n",
       "      <td>-0.038199</td>\n",
       "      <td>-0.06480</td>\n",
       "      <td>-0.003480</td>\n",
       "      <td>0.033500</td>\n",
       "      <td>-0.013467</td>\n",
       "      <td>0.08060</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030380</td>\n",
       "      <td>-0.102450</td>\n",
       "      <td>0.0841</td>\n",
       "      <td>-0.44800</td>\n",
       "      <td>0.11162</td>\n",
       "      <td>-0.114120</td>\n",
       "      <td>-0.195591</td>\n",
       "      <td>0.196667</td>\n",
       "      <td>-0.077160</td>\n",
       "      <td>-0.095075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 12670 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer    A1BG       A2M     A2ML1    A4GALT  \\\n",
       "GSM540550                    1.0  0.2610 -0.071700 -0.056388  0.115900   \n",
       "GSM540551                    1.0  0.0237 -0.075717 -0.025460  0.056557   \n",
       "GSM540552                    1.0  0.1170  0.184000  0.121440 -0.057271   \n",
       "GSM540553                    1.0  0.1540  0.176317  0.166813  0.139629   \n",
       "GSM540554                    1.0  0.1410  0.034750  0.002068 -0.060130   \n",
       "...                          ...     ...       ...       ...       ...   \n",
       "GSM540606                    0.0 -0.0340  0.115183 -0.013030 -0.120486   \n",
       "GSM540607                    0.0  0.0540  0.112083 -0.029762 -0.127914   \n",
       "GSM540608                    0.0 -0.1540  0.168233  0.093887  0.012549   \n",
       "GSM540609                    0.0  0.2420  0.139317  0.101675  0.150286   \n",
       "GSM540610                    0.0  0.0741  0.074317 -0.023225 -0.038199   \n",
       "\n",
       "             A4GNT      AAA1      AAAS      AACS    AADAC  ...      ZW10  \\\n",
       "GSM540550  0.18550 -0.074332  0.284667 -0.050671  0.14100  ... -0.011806   \n",
       "GSM540551 -0.09625  0.058537  0.048833 -0.124633 -0.00876  ...  0.074980   \n",
       "GSM540552 -0.28900  0.049406  0.168933  0.118744 -0.20600  ...  0.117400   \n",
       "GSM540553 -0.03006 -0.036533  0.314667  0.071756  0.12700  ...  0.120940   \n",
       "GSM540554 -0.06370  0.322153  0.049267 -0.056361  0.14400  ...  0.015800   \n",
       "...            ...       ...       ...       ...      ...  ...       ...   \n",
       "GSM540606  0.01380  0.096020 -0.082280 -0.084410  0.14000  ...  0.053585   \n",
       "GSM540607  0.00625  0.126520 -0.074100 -0.090500  0.15000  ... -0.000366   \n",
       "GSM540608 -0.11460 -0.061250  0.163033  0.201889  0.07380  ...  0.075702   \n",
       "GSM540609 -0.02901  0.000785  0.065833 -0.030744  0.07190  ...  0.066480   \n",
       "GSM540610 -0.06480 -0.003480  0.033500 -0.013467  0.08060  ... -0.030380   \n",
       "\n",
       "             ZWILCH   ZWINT     ZXDB     ZXDC    ZYG11A    ZYG11B       ZYX  \\\n",
       "GSM540550 -0.069583  0.0818 -0.26000  0.19156  0.021000  0.018978 -0.034700   \n",
       "GSM540551  0.105617  0.1890 -0.39000 -0.02402 -0.025660 -0.048632  0.021967   \n",
       "GSM540552  0.014768  0.1560 -0.51600 -0.26312 -0.214200 -0.240545  0.089867   \n",
       "GSM540553  0.090387  0.0823 -0.42100  0.12542 -0.152360 -0.211436  0.084667   \n",
       "GSM540554  0.022417  0.0821 -0.58000  0.07364 -0.020380 -0.014087 -0.098300   \n",
       "...             ...     ...      ...      ...       ...       ...       ...   \n",
       "GSM540606  0.004537  0.0128  0.00522 -0.05314 -0.080224 -0.110373  0.057200   \n",
       "GSM540607 -0.128767  0.0391  0.53800 -0.00796 -0.191920 -0.230038  0.041467   \n",
       "GSM540608  0.154717  0.1470 -0.50600  0.14348 -0.193660 -0.178318  0.072733   \n",
       "GSM540609 -0.007978  0.1480 -0.48600  0.05238 -0.168400 -0.209545  0.206333   \n",
       "GSM540610 -0.102450  0.0841 -0.44800  0.11162 -0.114120 -0.195591  0.196667   \n",
       "\n",
       "              ZZEF1      ZZZ3  \n",
       "GSM540550 -0.129433 -0.079166  \n",
       "GSM540551 -0.158727 -0.043763  \n",
       "GSM540552 -0.175639 -0.262375  \n",
       "GSM540553 -0.123181 -0.159617  \n",
       "GSM540554 -0.021133  0.018503  \n",
       "...             ...       ...  \n",
       "GSM540606 -0.026909 -0.087575  \n",
       "GSM540607 -0.100020 -0.126593  \n",
       "GSM540608  0.083072 -0.300383  \n",
       "GSM540609 -0.234480 -0.298500  \n",
       "GSM540610 -0.077160 -0.095075  \n",
       "\n",
       "[61 rows x 12670 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 61 samples.\n",
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 29 occurrences. This represents 47.54% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™æ®µä»£ç å®šä¹‰äº†ä¸€ä¸ªåä¸ºline_generatorçš„ç”Ÿæˆå™¨å‡½æ•°ï¼Œå®ƒçš„åŠŸèƒ½æ˜¯é€è¡Œè¯»å–æ¥è‡ªæ–‡ä»¶æˆ–å­—ç¬¦ä¸²çš„å†…å®¹ï¼Œå¹¶é€è¡Œäº§ç”Ÿå¤„ç†è¿‡çš„æ•°æ®ï¼ˆåŽ»é™¤äº†æ¯è¡Œæœ«å°¾çš„ç©ºç™½å­—ç¬¦ï¼‰ã€‚\n",
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™ä¸ªå‡½æ•°filter_content_by_prefixçš„ç›®çš„æ˜¯ä»Žä¸€ä¸ªæ–‡ä»¶æˆ–ä¸€ä¸²æ–‡æœ¬ä¸­è¿‡æ»¤å‡ºæ»¡è¶³ç‰¹å®šå‰ç¼€æ¡ä»¶çš„è¡Œã€‚\n",
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™ä¸ªget_background_and_clinical_dataå‡½æ•°æ—¨åœ¨ä»Žä¸€ä¸ªç‰¹å®šæ ¼å¼çš„æ–‡ä»¶ä¸­æå–æ•°æ®é›†çš„èƒŒæ™¯ä¿¡æ¯å’Œæ ·æœ¬ç‰¹å¾æ•°æ®ã€‚\n",
    "# è¿™ä¸ªå‡½æ•°ä½¿ç”¨å‰é¢å®šä¹‰çš„filter_content_by_prefixå‡½æ•°æ¥ç­›é€‰å‡ºæ–‡ä»¶ä¸­æ»¡è¶³æŒ‡å®šå‰ç¼€æ¡ä»¶çš„è¡Œã€‚\n",
    "\n",
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Advancing a Clinically Relevant Perspective of the Clonal Nature of Cancer\"\n",
      "!Series_summary\t\"We used DNA content-based flow cytometry to distinguish and isolate nuclei from clonal populations in primary tissues from three disparate cancers with variable clinical histories. We then developed a methodology to adapt flow cytometrically purified nuclei samples for use with whole genome technologies including aCGH and next generation sequencing (NGS). Our results demonstrate that selected aberrations in the genomes of distinct clonal populations in each patient create clinically relevant contexts at least with respect to the cancer types profiled in this study.\"\n",
      "!Series_overall_design\t\"We applied DNA content based flow sorting to isolate the nuclei of clonal populations from tumor biopsies. Genomic DNA from each sorted population was amplified with phi29 polymerase. A 1ug aliquot of each amplified sample was digested with DNAse 1 then labeled with Cy5 using a Klenow-based commercial kit (Invitrogen). Each sample was hybridized with a pooled normal (46,XX) reference (Promega) to Agilent 244k CGH arrays. The use of highly purified objectively defined flow sorted populations provides high definition genomic profiles of clonal populations from pancreatic adenocarcinomas (PA), adrenal cortical carcinomas (ACC), and prostate adenocarcinomas (PC).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                 GSM540550  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540551  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540552  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540553  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540554  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540555  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540556  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540557  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540558  ...  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma  ...   \n",
       "\n",
       "                            GSM540601                           GSM540602  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Adrenal Cortical Carcinoma   \n",
       "\n",
       "                            GSM540603                   GSM540604  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540605                   GSM540606  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540607                   GSM540608  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540609                   GSM540610  \n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                 GSM540550  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540551  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540552  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540553  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540554  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540555  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540556  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540557  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540558  ...  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma  ...   \n",
       "\n",
       "                            GSM540601                           GSM540602  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Adrenal Cortical Carcinoma   \n",
       "\n",
       "                            GSM540603                   GSM540604  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540605                   GSM540606  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540607                   GSM540608  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540609                   GSM540610  \n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                 GSM540550  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540551  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540552  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540553  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540554  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540555  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540556  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540557  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540558  ...  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma  ...   \n",
       "\n",
       "                            GSM540601                           GSM540602  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Adrenal Cortical Carcinoma   \n",
       "\n",
       "                            GSM540603                   GSM540604  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540605                   GSM540606  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540607                   GSM540608  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540609                   GSM540610  \n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PandasGUI INFO â€” pandasgui.gui â€” Opening PandasGUI\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\4289982574.py:1: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\4289982574.py:1: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\4289982574.py:1: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\4289982574.py:1: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\4289982574.py:1: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\4289982574.py:1: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pandasgui.gui.PandasGui at 0x1f2304ac160>"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "show(clinical_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™ä¸ªget_unique_values_by_rowå‡½æ•°çš„ç›®çš„æ˜¯ä¸ºäº†ä»Žç»™å®šçš„pandas DataFrameä¸­ï¼ŒæŒ‰è¡Œæ•´ç†æ¯è¡Œä¸­çš„å”¯ä¸€å€¼ï¼Œå¹¶å°†è¿™äº›å”¯ä¸€å€¼ä»¥å­—å…¸çš„å½¢å¼ç»„ç»‡èµ·æ¥ã€‚\n",
    "# è¿™ä¸ªå­—å…¸çš„é”®æ˜¯DataFrameçš„è¡Œç´¢å¼•ï¼Œè€Œå€¼æ˜¯ä¸€ä¸ªåŒ…å«äº†è¯¥è¡Œä¸­å”¯ä¸€å€¼çš„åˆ—è¡¨\n",
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['tissue: Pancreatic Ductal Adenocarcinoma',\n",
       "  'tissue: Adrenal Cortical Carcinoma',\n",
       "  'tissue: Prostate Carcinoma']}"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "è¿™ä¸ªå­—å…¸åªæ˜¾ç¤ºäº†ä¸€ä¸ªé”®å€¼å¯¹ï¼Œå¯èƒ½æ˜¯å› ä¸ºç¤ºä¾‹åªå±•ç¤ºäº†ä¸€ä¸ªç®€åŒ–çš„è¾“å‡ºï¼Œæˆ–è€…clinical_data DataFrameå®žé™…ä¸Šåªæœ‰ä¸€è¡Œæ•°æ®ã€‚åœ¨å®žé™…åº”ç”¨ä¸­ï¼Œå¦‚æžœDataFrameæœ‰å¤šè¡Œï¼Œè¿™ä¸ªå­—å…¸å°†åŒ…å«æ›´å¤šçš„é”®å€¼å¯¹ï¼Œæ¯ä¸ªé”®å€¼å¯¹å¯¹åº”ä¸€è¡Œä¸­çš„å”¯ä¸€å€¼åˆ—è¡¨ã€‚"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Adrenocortical Cancer\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Adrenocortical Cancer\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Advancing a Clinically Relevant Perspective of the Clonal Nature of Cancer\"\\n!Series_summary\\t\"We used DNA content-based flow cytometry to distinguish and isolate nuclei from clonal populations in primary tissues from three disparate cancers with variable clinical histories. We then developed a methodology to adapt flow cytometrically purified nuclei samples for use with whole genome technologies including aCGH and next generation sequencing (NGS). Our results demonstrate that selected aberrations in the genomes of distinct clonal populations in each patient create clinically relevant contexts at least with respect to the cancer types profiled in this study.\"\\n!Series_overall_design\\t\"We applied DNA content based flow sorting to isolate the nuclei of clonal populations from tumor biopsies. Genomic DNA from each sorted population was amplified with phi29 polymerase. A 1ug aliquot of each amplified sample was digested with DNAse 1 then labeled with Cy5 using a Klenow-based commercial kit (Invitrogen). Each sample was hybridized with a pooled normal (46,XX) reference (Promega) to Agilent 244k CGH arrays. The use of highly purified objectively defined flow sorted populations provides high definition genomic profiles of clonal populations from pancreatic adenocarcinomas (PA), adrenal cortical carcinomas (ACC), and prostate adenocarcinomas (PC).\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'tissue: Pancreatic Ductal Adenocarcinoma\\', \\'tissue: Adrenal Cortical Carcinoma\\', \\'tissue: Prostate Carcinoma\\']}\\n'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Pancreatic Ductal Adenocarcinoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™ä¸ªget_feature_dataå‡½æ•°çš„ç›®çš„æ˜¯ä»Žä¸€ä¸ªæ ·æœ¬ç‰¹å¾çš„DataFrameä¸­æå–ç‰¹å®šç‰¹å¾çš„æ•°æ®ï¼Œå¹¶ä½¿ç”¨ç»™å®šçš„è½¬æ¢å‡½æ•°å°†è¿™äº›æ•°æ®è½¬æ¢ä¸ºäºŒè¿›åˆ¶æˆ–è¿žç»­å˜é‡çš„æ ¼å¼ã€‚\n",
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™ä¸ªgeo_select_clinical_featureså‡½æ•°æ˜¯ä¸ºäº†ä»ŽGEOï¼ˆGene Expression Omnibusï¼‰æ•°æ®åº“ç³»åˆ—ä¸­ä»£è¡¨æ ·æœ¬ç‰¹å¾çš„DataFrameæå–å¹¶å¤„ç†ç‰¹å®šçš„ä¸´åºŠç‰¹å¾ã€‚\n",
    "# å‡½æ•°é€šè¿‡é›†æˆä¸åŒçš„ç‰¹å¾å¤„ç†æµç¨‹ï¼Œä½¿å¾—ä»Žå¤æ‚çš„æ•°æ®é›†ä¸­æå–ã€è½¬æ¢å’Œæ•´åˆç‰¹å®šç‰¹å¾å˜å¾—ç®€å•ç›´æŽ¥ã€‚\n",
    "\n",
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>tissue: Pancreatic Ductal Adenocarcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Adrenal Cortical Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "      <td>tissue: Prostate Carcinoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                 GSM540550  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540551  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540552  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540553  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540554  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540555  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540556  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540557  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma   \n",
       "\n",
       "                                  GSM540558  ...  \\\n",
       "0  tissue: Pancreatic Ductal Adenocarcinoma  ...   \n",
       "\n",
       "                            GSM540601                           GSM540602  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Adrenal Cortical Carcinoma   \n",
       "\n",
       "                            GSM540603                   GSM540604  \\\n",
       "0  tissue: Adrenal Cortical Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540605                   GSM540606  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540607                   GSM540608  \\\n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma   \n",
       "\n",
       "                    GSM540609                   GSM540610  \n",
       "0  tissue: Prostate Carcinoma  tissue: Prostate Carcinoma  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\882291112.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\882291112.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Length mismatch: Expected axis has 0 elements, new values have 1 elements",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[80], line 8\u001b[0m\n\u001b[0;32m      4\u001b[0m gender_row \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m      6\u001b[0m trait_type \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbinary\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m----> 8\u001b[0m selected_clinical_data \u001b[38;5;241m=\u001b[39m \u001b[43mgeo_select_clinical_features\u001b[49m\u001b[43m(\u001b[49m\u001b[43mclinical_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mTRAIT\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtrait_row\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert_trait\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mage_row\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mage_row\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m      9\u001b[0m \u001b[43m                                                      \u001b[49m\u001b[43mconvert_age\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_age\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgender_row\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgender_row\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     10\u001b[0m \u001b[43m                                                      \u001b[49m\u001b[43mconvert_gender\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconvert_gender\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     11\u001b[0m selected_clinical_data\u001b[38;5;241m.\u001b[39mhead()\n",
      "Cell \u001b[1;32mIn[77], line 35\u001b[0m, in \u001b[0;36mgeo_select_clinical_features\u001b[1;34m(clinical_df, trait, trait_row, convert_trait, age_row, convert_age, gender_row, convert_gender)\u001b[0m\n\u001b[0;32m     33\u001b[0m     feature_list\u001b[38;5;241m.\u001b[39mappend(age_data)\n\u001b[0;32m     34\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m gender_row \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m---> 35\u001b[0m     gender_data \u001b[38;5;241m=\u001b[39m \u001b[43mget_feature_data\u001b[49m\u001b[43m(\u001b[49m\u001b[43mclinical_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgender_row\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mGender\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert_gender\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     36\u001b[0m     feature_list\u001b[38;5;241m.\u001b[39mappend(gender_data)\n\u001b[0;32m     38\u001b[0m selected_clinical_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat(feature_list, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n",
      "Cell \u001b[1;32mIn[76], line 6\u001b[0m, in \u001b[0;36mget_feature_data\u001b[1;34m(clinical_df, row_id, feature, convert_fn)\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\u001b[39;00m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;124;03ma binary or continuous variable\"\"\"\u001b[39;00m\n\u001b[0;32m      5\u001b[0m clinical_df \u001b[38;5;241m=\u001b[39m clinical_df\u001b[38;5;241m.\u001b[39miloc[row_id:row_id \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39mdrop(columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!Sample_geo_accession\u001b[39m\u001b[38;5;124m'\u001b[39m], errors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m----> 6\u001b[0m \u001b[43mclinical_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m \u001b[38;5;241m=\u001b[39m [feature]\n\u001b[0;32m      7\u001b[0m clinical_df \u001b[38;5;241m=\u001b[39m clinical_df\u001b[38;5;241m.\u001b[39mapplymap(convert_fn)\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m clinical_df\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\generic.py:6310\u001b[0m, in \u001b[0;36mNDFrame.__setattr__\u001b[1;34m(self, name, value)\u001b[0m\n\u001b[0;32m   6308\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   6309\u001b[0m     \u001b[38;5;28mobject\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__getattribute__\u001b[39m(\u001b[38;5;28mself\u001b[39m, name)\n\u001b[1;32m-> 6310\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mobject\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__setattr__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   6311\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m:\n\u001b[0;32m   6312\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n",
      "File \u001b[1;32mproperties.pyx:69\u001b[0m, in \u001b[0;36mpandas._libs.properties.AxisProperty.__set__\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\generic.py:813\u001b[0m, in \u001b[0;36mNDFrame._set_axis\u001b[1;34m(self, axis, labels)\u001b[0m\n\u001b[0;32m    808\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    809\u001b[0m \u001b[38;5;124;03mThis is called from the cython code when we set the `index` attribute\u001b[39;00m\n\u001b[0;32m    810\u001b[0m \u001b[38;5;124;03mdirectly, e.g. `series.index = [1, 2, 3]`.\u001b[39;00m\n\u001b[0;32m    811\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    812\u001b[0m labels \u001b[38;5;241m=\u001b[39m ensure_index(labels)\n\u001b[1;32m--> 813\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mset_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlabels\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    814\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_clear_item_cache()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\internals\\managers.py:238\u001b[0m, in \u001b[0;36mBaseBlockManager.set_axis\u001b[1;34m(self, axis, new_labels)\u001b[0m\n\u001b[0;32m    236\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mset_axis\u001b[39m(\u001b[38;5;28mself\u001b[39m, axis: AxisInt, new_labels: Index) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    237\u001b[0m     \u001b[38;5;66;03m# Caller is responsible for ensuring we have an Index object.\u001b[39;00m\n\u001b[1;32m--> 238\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_set_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnew_labels\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    239\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes[axis] \u001b[38;5;241m=\u001b[39m new_labels\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\internals\\base.py:98\u001b[0m, in \u001b[0;36mDataManager._validate_set_axis\u001b[1;34m(self, axis, new_labels)\u001b[0m\n\u001b[0;32m     95\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[0;32m     97\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m new_len \u001b[38;5;241m!=\u001b[39m old_len:\n\u001b[1;32m---> 98\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m     99\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLength mismatch: Expected axis has \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mold_len\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m elements, new \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    100\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalues have \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnew_len\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m elements\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    101\u001b[0m     )\n",
      "\u001b[1;31mValueError\u001b[0m: Length mismatch: Expected axis has 0 elements, new values have 1 elements"
     ]
    }
   ],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = 0\n",
    "gender_row = 1\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PandasGUI INFO â€” pandasgui.gui â€” Opening PandasGUI\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\915109071.py:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  show(selected_clinical_data)\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\915109071.py:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  show(selected_clinical_data)\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\915109071.py:3: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\915109071.py:3: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\915109071.py:3: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_15160\\915109071.py:3: FutureWarning:\n",
      "\n",
      "Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pandasgui.gui.PandasGui at 0x1f221489ea0>"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pandasgui import show\n",
    "\n",
    "show(selected_clinical_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# è¿™ä¸ªå‡½æ•°get_genetic_dataæ˜¯ç”¨æ¥è¯»å–åŸºå› è¡¨è¾¾æ•°æ®æ–‡ä»¶ï¼Œå¹¶å°†å…¶è½¬æ¢ä¸ºpandas DataFrameå¯¹è±¡çš„æ ¼å¼ï¼ŒåŒæ—¶è¿›è¡Œä¸€äº›æ ¼å¼è°ƒæ•´ã€‚\n",
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.00599</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>-0.0779</td>\n",
       "      <td>0.01020</td>\n",
       "      <td>-0.00558</td>\n",
       "      <td>-0.0871</td>\n",
       "      <td>-0.0895</td>\n",
       "      <td>-0.114</td>\n",
       "      <td>-0.00695</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0563</td>\n",
       "      <td>0.00207</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.00841</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.0128</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>-0.000229</td>\n",
       "      <td>-0.0148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1730</td>\n",
       "      <td>-0.43800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.6140</td>\n",
       "      <td>-0.019</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.5670</td>\n",
       "      <td>-0.004620</td>\n",
       "      <td>-0.4700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.2000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3190</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.4420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.6850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.19900</td>\n",
       "      <td>0.1570</td>\n",
       "      <td>0.2230</td>\n",
       "      <td>0.01230</td>\n",
       "      <td>0.13200</td>\n",
       "      <td>0.0682</td>\n",
       "      <td>0.1330</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.03560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1920</td>\n",
       "      <td>0.3650</td>\n",
       "      <td>0.36400</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>-0.24800</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>-0.0865</td>\n",
       "      <td>0.0336</td>\n",
       "      <td>0.167000</td>\n",
       "      <td>0.0337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.1050</td>\n",
       "      <td>-0.06010</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.1780</td>\n",
       "      <td>-0.00527</td>\n",
       "      <td>0.00652</td>\n",
       "      <td>-0.1610</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>-0.152</td>\n",
       "      <td>-0.02830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0991</td>\n",
       "      <td>-0.0738</td>\n",
       "      <td>-0.05840</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>-0.06350</td>\n",
       "      <td>-0.1470</td>\n",
       "      <td>-0.2470</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>-0.115000</td>\n",
       "      <td>0.0135</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "ID                                                                     \n",
       "1      0.0201    0.00599     0.0117    -0.0779    0.01020   -0.00558   \n",
       "2      0.0000    0.00000     0.0000    -0.1730   -0.43800    0.00000   \n",
       "3     -0.2000    0.00000     0.0000     0.0000    0.00000    0.00000   \n",
       "4      0.1210    0.19900     0.1570     0.2230    0.01230    0.13200   \n",
       "5     -0.1050   -0.06010    -0.1140    -0.1780   -0.00527    0.00652   \n",
       "\n",
       "    GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  GSM540602  \\\n",
       "ID                                              ...                         \n",
       "1     -0.0871    -0.0895     -0.114   -0.00695  ...     0.0592     0.0563   \n",
       "2      0.0000    -0.6140     -0.019    0.00000  ...     0.0000     0.0000   \n",
       "3      0.0000    -0.3190      0.000    0.00000  ...     0.0000     0.0000   \n",
       "4      0.0682     0.1330      0.148    0.03560  ...     0.1920     0.3650   \n",
       "5     -0.1610    -0.1200     -0.152   -0.02830  ...     0.0991    -0.0738   \n",
       "\n",
       "    GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  GSM540608  \\\n",
       "ID                                                                     \n",
       "1     0.00207     0.0208    0.00841     0.0122    -0.0128     0.0748   \n",
       "2     0.00000     0.0000    0.00000     0.0000     0.0000    -0.5670   \n",
       "3     0.00000     0.0000    0.00000     0.0000     0.0000    -0.4420   \n",
       "4     0.36400     0.0321   -0.24800     0.0240    -0.0865     0.0336   \n",
       "5    -0.05840     0.0120   -0.06350    -0.1470    -0.2470     0.1170   \n",
       "\n",
       "    GSM540609  GSM540610  \n",
       "ID                        \n",
       "1   -0.000229    -0.0148  \n",
       "2   -0.004620    -0.4700  \n",
       "3    0.000000    -0.6850  \n",
       "4    0.167000     0.0337  \n",
       "5   -0.115000     0.0135  \n",
       "\n",
       "[5 rows x 61 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM540550</th>\n",
       "      <th>GSM540551</th>\n",
       "      <th>GSM540552</th>\n",
       "      <th>GSM540553</th>\n",
       "      <th>GSM540554</th>\n",
       "      <th>GSM540555</th>\n",
       "      <th>GSM540556</th>\n",
       "      <th>GSM540557</th>\n",
       "      <th>GSM540558</th>\n",
       "      <th>GSM540559</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM540601</th>\n",
       "      <th>GSM540602</th>\n",
       "      <th>GSM540603</th>\n",
       "      <th>GSM540604</th>\n",
       "      <th>GSM540605</th>\n",
       "      <th>GSM540606</th>\n",
       "      <th>GSM540607</th>\n",
       "      <th>GSM540608</th>\n",
       "      <th>GSM540609</th>\n",
       "      <th>GSM540610</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0201</td>\n",
       "      <td>0.00599</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>-0.0779</td>\n",
       "      <td>0.01020</td>\n",
       "      <td>-0.00558</td>\n",
       "      <td>-0.0871</td>\n",
       "      <td>-0.0895</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.00695</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0563</td>\n",
       "      <td>0.002070</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.00841</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.01280</td>\n",
       "      <td>0.0748</td>\n",
       "      <td>-0.000229</td>\n",
       "      <td>-0.0148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.1730</td>\n",
       "      <td>-0.43800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.6140</td>\n",
       "      <td>-0.0190</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.5670</td>\n",
       "      <td>-0.004620</td>\n",
       "      <td>-0.4700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.2000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.3190</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.4420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.6850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.19900</td>\n",
       "      <td>0.1570</td>\n",
       "      <td>0.2230</td>\n",
       "      <td>0.01230</td>\n",
       "      <td>0.13200</td>\n",
       "      <td>0.0682</td>\n",
       "      <td>0.1330</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>0.03560</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1920</td>\n",
       "      <td>0.3650</td>\n",
       "      <td>0.364000</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>-0.24800</td>\n",
       "      <td>0.0240</td>\n",
       "      <td>-0.08650</td>\n",
       "      <td>0.0336</td>\n",
       "      <td>0.167000</td>\n",
       "      <td>0.0337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.1050</td>\n",
       "      <td>-0.06010</td>\n",
       "      <td>-0.1140</td>\n",
       "      <td>-0.1780</td>\n",
       "      <td>-0.00527</td>\n",
       "      <td>0.00652</td>\n",
       "      <td>-0.1610</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>-0.1520</td>\n",
       "      <td>-0.02830</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0991</td>\n",
       "      <td>-0.0738</td>\n",
       "      <td>-0.058400</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>-0.06350</td>\n",
       "      <td>-0.1470</td>\n",
       "      <td>-0.24700</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>-0.115000</td>\n",
       "      <td>0.0135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243500</th>\n",
       "      <td>-0.2810</td>\n",
       "      <td>-0.34800</td>\n",
       "      <td>-0.3390</td>\n",
       "      <td>-0.2320</td>\n",
       "      <td>0.02470</td>\n",
       "      <td>0.02690</td>\n",
       "      <td>-0.2440</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>-0.1790</td>\n",
       "      <td>-0.34600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>-0.1520</td>\n",
       "      <td>-0.211000</td>\n",
       "      <td>-0.0559</td>\n",
       "      <td>0.08870</td>\n",
       "      <td>-0.0608</td>\n",
       "      <td>0.00423</td>\n",
       "      <td>-0.4000</td>\n",
       "      <td>-0.069500</td>\n",
       "      <td>-0.0305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243501</th>\n",
       "      <td>-0.1770</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.2670</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.28800</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0433</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2160</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.122000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243502</th>\n",
       "      <td>-0.5300</td>\n",
       "      <td>1.30000</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.66700</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-0.2540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.1480</td>\n",
       "      <td>-0.302000</td>\n",
       "      <td>-0.9700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243503</th>\n",
       "      <td>0.0344</td>\n",
       "      <td>0.02240</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>-0.0692</td>\n",
       "      <td>0.03500</td>\n",
       "      <td>0.01120</td>\n",
       "      <td>-0.0938</td>\n",
       "      <td>-0.0330</td>\n",
       "      <td>-0.1030</td>\n",
       "      <td>0.00613</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0709</td>\n",
       "      <td>0.0626</td>\n",
       "      <td>-0.000328</td>\n",
       "      <td>0.0182</td>\n",
       "      <td>0.01140</td>\n",
       "      <td>0.0342</td>\n",
       "      <td>-0.03030</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>-0.0138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243504</th>\n",
       "      <td>0.0434</td>\n",
       "      <td>0.01580</td>\n",
       "      <td>0.0122</td>\n",
       "      <td>-0.0646</td>\n",
       "      <td>0.05200</td>\n",
       "      <td>0.00951</td>\n",
       "      <td>-0.0797</td>\n",
       "      <td>-0.0113</td>\n",
       "      <td>-0.1080</td>\n",
       "      <td>-0.00240</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0639</td>\n",
       "      <td>0.0483</td>\n",
       "      <td>-0.002440</td>\n",
       "      <td>0.0539</td>\n",
       "      <td>0.02960</td>\n",
       "      <td>0.0264</td>\n",
       "      <td>-0.00543</td>\n",
       "      <td>0.0700</td>\n",
       "      <td>-0.012400</td>\n",
       "      <td>-0.0203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243429 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM540550  GSM540551  GSM540552  GSM540553  GSM540554  GSM540555  \\\n",
       "ID                                                                         \n",
       "1          0.0201    0.00599     0.0117    -0.0779    0.01020   -0.00558   \n",
       "2          0.0000    0.00000     0.0000    -0.1730   -0.43800    0.00000   \n",
       "3         -0.2000    0.00000     0.0000     0.0000    0.00000    0.00000   \n",
       "4          0.1210    0.19900     0.1570     0.2230    0.01230    0.13200   \n",
       "5         -0.1050   -0.06010    -0.1140    -0.1780   -0.00527    0.00652   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "243500    -0.2810   -0.34800    -0.3390    -0.2320    0.02470    0.02690   \n",
       "243501    -0.1770    0.00000    -0.2670     0.0000   -0.28800    0.00000   \n",
       "243502    -0.5300    1.30000     0.0179     0.0000   -0.66700    0.00000   \n",
       "243503     0.0344    0.02240     0.0275    -0.0692    0.03500    0.01120   \n",
       "243504     0.0434    0.01580     0.0122    -0.0646    0.05200    0.00951   \n",
       "\n",
       "        GSM540556  GSM540557  GSM540558  GSM540559  ...  GSM540601  GSM540602  \\\n",
       "ID                                                  ...                         \n",
       "1         -0.0871    -0.0895    -0.1140   -0.00695  ...     0.0592     0.0563   \n",
       "2          0.0000    -0.6140    -0.0190    0.00000  ...     0.0000     0.0000   \n",
       "3          0.0000    -0.3190     0.0000    0.00000  ...     0.0000     0.0000   \n",
       "4          0.0682     0.1330     0.1480    0.03560  ...     0.1920     0.3650   \n",
       "5         -0.1610    -0.1200    -0.1520   -0.02830  ...     0.0991    -0.0738   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "243500    -0.2440     0.0160    -0.1790   -0.34600  ...     0.1480    -0.1520   \n",
       "243501     0.0000     0.0000     0.0433    0.00000  ...     0.2160     0.0000   \n",
       "243502     0.0000     0.0000     0.0000    0.00000  ...     0.0000    -0.2540   \n",
       "243503    -0.0938    -0.0330    -0.1030    0.00613  ...     0.0709     0.0626   \n",
       "243504    -0.0797    -0.0113    -0.1080   -0.00240  ...     0.0639     0.0483   \n",
       "\n",
       "        GSM540603  GSM540604  GSM540605  GSM540606  GSM540607  GSM540608  \\\n",
       "ID                                                                         \n",
       "1        0.002070     0.0208    0.00841     0.0122   -0.01280     0.0748   \n",
       "2        0.000000     0.0000    0.00000     0.0000    0.00000    -0.5670   \n",
       "3        0.000000     0.0000    0.00000     0.0000    0.00000    -0.4420   \n",
       "4        0.364000     0.0321   -0.24800     0.0240   -0.08650     0.0336   \n",
       "5       -0.058400     0.0120   -0.06350    -0.1470   -0.24700     0.1170   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "243500  -0.211000    -0.0559    0.08870    -0.0608    0.00423    -0.4000   \n",
       "243501   0.122000     0.0000    0.00000     0.0000    0.00000     0.0000   \n",
       "243502   0.000000     0.0000    0.00000     0.0000    0.00000    -0.1480   \n",
       "243503  -0.000328     0.0182    0.01140     0.0342   -0.03030     0.0543   \n",
       "243504  -0.002440     0.0539    0.02960     0.0264   -0.00543     0.0700   \n",
       "\n",
       "        GSM540609  GSM540610  \n",
       "ID                            \n",
       "1       -0.000229    -0.0148  \n",
       "2       -0.004620    -0.4700  \n",
       "3        0.000000    -0.6850  \n",
       "4        0.167000     0.0337  \n",
       "5       -0.115000     0.0135  \n",
       "...           ...        ...  \n",
       "243500  -0.069500    -0.0305  \n",
       "243501   0.000000    -1.0400  \n",
       "243502  -0.302000    -0.9700  \n",
       "243503   0.014000    -0.0138  \n",
       "243504  -0.012400    -0.0203  \n",
       "\n",
       "[243429 rows x 61 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1',\n",
       " '2',\n",
       " '3',\n",
       " '4',\n",
       " '5',\n",
       " '6',\n",
       " '7',\n",
       " '8',\n",
       " '9',\n",
       " '10',\n",
       " '11',\n",
       " '13',\n",
       " '14',\n",
       " '15',\n",
       " '16',\n",
       " '17',\n",
       " '18',\n",
       " '19',\n",
       " '20',\n",
       " '21']"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# è¿™äº›ä»£ç å°†èŽ·å–DataFrameç´¢å¼•çš„å‰20ä¸ªå…ƒç´ ï¼Œå¹¶å°†å®ƒä»¬è½¬æ¢ä¸ºåˆ—è¡¨ã€‚åœ¨è¿™ä¸ªä¸Šä¸‹æ–‡ä¸­ï¼Œè¿™ä¸ªç´¢å¼•ä»£è¡¨åŸºå› çš„IDã€‚å› æ­¤ï¼Œgene_row_idså°†åŒ…å«å‰20ä¸ªåŸºå› çš„IDã€‚\n",
    "# å¦‚æžœæ‚¨åœ¨Jupyter Notebookæˆ–ç±»ä¼¼çŽ¯å¢ƒä¸­è¿è¡Œè¿™äº›ä»£ç ï¼Œå®ƒä»¬å°†æ‰“å°å‡ºDataFrameçš„å‰å‡ è¡Œï¼Œä»¥åŠä¸€ä¸ªåŒ…å«å‰20ä¸ªåŸºå› IDçš„åˆ—è¡¨ã€‚\n",
    "# è¿™å¯ä»¥å¸®åŠ©æ‚¨äº†è§£å“ªäº›åŸºå› åŒ…å«åœ¨æ•°æ®é›†çš„é¡¶éƒ¨ï¼Œå¹¶ä¸”å¯¹è¿›è¡ŒåŽç»­åˆ†æžï¼Œæ¯”å¦‚ç‰¹å®šåŸºå› çš„è¡¨è¾¾é‡åˆ†æžæä¾›ä¸€ä¸ªèµ·ç‚¹ã€‚\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '13', '14', '15', '16', '17', '18', '19', '20', '21']\\n\""
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1', '2', '3', '4', '5'], 'COL': ['267', '267', '267', '267', '267'], 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0], 'SPOT_ID': ['HsCGHBrightCorner', 'DarkCorner', 'DarkCorner', 'A_16_P20527812', 'A_16_P01708709'], 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'], 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'], 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'], 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'], 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'], 'CHROMOSOMAL_LOCATION': [nan, nan, nan, 'chr16:076331867-076331926', 'chr7:047626734-047626793'], 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'], 'DESCRIPTION': [nan, nan, nan, nan, 'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'], 'GB_RANGE': [nan, nan, nan, 'NC_000016.8[076331867..076331926]', 'NC_000007.11[047626734..047626793]']}\n"
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '13', '14', '15', '16', '17', '18', '19', '20', '21']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1', '2', '3', '4', '5'], 'COL': ['267', '267', '267', '267', '267'], 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0], 'SPOT_ID': ['HsCGHBrightCorner', 'DarkCorner', 'DarkCorner', 'A_16_P20527812', 'A_16_P01708709'], 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'], 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'], 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'], 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'], 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'], 'CHROMOSOMAL_LOCATION': [nan, nan, nan, 'chr16:076331867-076331926', 'chr7:047626734-047626793'], 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'], 'DESCRIPTION': [nan, nan, nan, nan, 'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'], 'GB_RANGE': [nan, nan, nan, 'NC_000016.8[076331867..076331926]', 'NC_000007.11[047626734..047626793]']}\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ID': ['1', '2', '3', '4', '5'],\n",
       " 'COL': ['267', '267', '267', '267', '267'],\n",
       " 'ROW': [912.0, 910.0, 908.0, 906.0, 904.0],\n",
       " 'SPOT_ID': ['HsCGHBrightCorner',\n",
       "  'DarkCorner',\n",
       "  'DarkCorner',\n",
       "  'A_16_P20527812',\n",
       "  'A_16_P01708709'],\n",
       " 'CONTROL_TYPE': ['pos', 'pos', 'pos', 'FALSE', 'FALSE'],\n",
       " 'GB_ACC': [nan, nan, nan, nan, 'NM_138295'],\n",
       " 'GENE_SYMBOL': [nan, nan, nan, nan, 'PKD1L1'],\n",
       " 'GENE_NAME': [nan, nan, nan, nan, 'polycystic kidney disease 1 like 1'],\n",
       " 'ACCESSION_STRING': [nan, nan, nan, nan, 'ref|NM_138295|ref|NM_025031'],\n",
       " 'CHROMOSOMAL_LOCATION': [nan,\n",
       "  nan,\n",
       "  nan,\n",
       "  'chr16:076331867-076331926',\n",
       "  'chr7:047626734-047626793'],\n",
       " 'CYTOBAND': [nan, nan, nan, 'hs|q23.1', 'hs|p12.3'],\n",
       " 'DESCRIPTION': [nan,\n",
       "  nan,\n",
       "  nan,\n",
       "  nan,\n",
       "  'Homo sapiens polycystic kidney disease 1 like 1 (PKD1L1), mRNA.'],\n",
       " 'GB_RANGE': [nan,\n",
       "  nan,\n",
       "  nan,\n",
       "  'NC_000016.8[076331867..076331926]',\n",
       "  'NC_000007.11[047626734..047626793]']}"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>COL</th>\n",
       "      <th>ROW</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>CONTROL_TYPE</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ACCESSION_STRING</th>\n",
       "      <th>CHROMOSOMAL_LOCATION</th>\n",
       "      <th>CYTOBAND</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>GB_RANGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>267</td>\n",
       "      <td>912.0</td>\n",
       "      <td>HsCGHBrightCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>267</td>\n",
       "      <td>910.0</td>\n",
       "      <td>DarkCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>267</td>\n",
       "      <td>908.0</td>\n",
       "      <td>DarkCorner</td>\n",
       "      <td>pos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>267</td>\n",
       "      <td>906.0</td>\n",
       "      <td>A_16_P20527812</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr16:076331867-076331926</td>\n",
       "      <td>hs|q23.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000016.8[076331867..076331926]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>267</td>\n",
       "      <td>904.0</td>\n",
       "      <td>A_16_P01708709</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>NM_138295</td>\n",
       "      <td>PKD1L1</td>\n",
       "      <td>polycystic kidney disease 1 like 1</td>\n",
       "      <td>ref|NM_138295|ref|NM_025031</td>\n",
       "      <td>chr7:047626734-047626793</td>\n",
       "      <td>hs|p12.3</td>\n",
       "      <td>Homo sapiens polycystic kidney disease 1 like ...</td>\n",
       "      <td>NC_000007.11[047626734..047626793]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092790</th>\n",
       "      <td>243500</td>\n",
       "      <td>-3.05E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092791</th>\n",
       "      <td>243501</td>\n",
       "      <td>-1.04E+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092792</th>\n",
       "      <td>243502</td>\n",
       "      <td>-9.70E-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092793</th>\n",
       "      <td>243503</td>\n",
       "      <td>-1.38E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15092794</th>\n",
       "      <td>243504</td>\n",
       "      <td>-2.03E-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15092795 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID        COL    ROW            SPOT_ID CONTROL_TYPE     GB_ACC  \\\n",
       "0              1        267  912.0  HsCGHBrightCorner          pos        NaN   \n",
       "1              2        267  910.0         DarkCorner          pos        NaN   \n",
       "2              3        267  908.0         DarkCorner          pos        NaN   \n",
       "3              4        267  906.0     A_16_P20527812        FALSE        NaN   \n",
       "4              5        267  904.0     A_16_P01708709        FALSE  NM_138295   \n",
       "...          ...        ...    ...                ...          ...        ...   \n",
       "15092790  243500  -3.05E-02    NaN                NaN          NaN        NaN   \n",
       "15092791  243501  -1.04E+00    NaN                NaN          NaN        NaN   \n",
       "15092792  243502  -9.70E-01    NaN                NaN          NaN        NaN   \n",
       "15092793  243503  -1.38E-02    NaN                NaN          NaN        NaN   \n",
       "15092794  243504  -2.03E-02    NaN                NaN          NaN        NaN   \n",
       "\n",
       "         GENE_SYMBOL                           GENE_NAME  \\\n",
       "0                NaN                                 NaN   \n",
       "1                NaN                                 NaN   \n",
       "2                NaN                                 NaN   \n",
       "3                NaN                                 NaN   \n",
       "4             PKD1L1  polycystic kidney disease 1 like 1   \n",
       "...              ...                                 ...   \n",
       "15092790         NaN                                 NaN   \n",
       "15092791         NaN                                 NaN   \n",
       "15092792         NaN                                 NaN   \n",
       "15092793         NaN                                 NaN   \n",
       "15092794         NaN                                 NaN   \n",
       "\n",
       "                     ACCESSION_STRING       CHROMOSOMAL_LOCATION  CYTOBAND  \\\n",
       "0                                 NaN                        NaN       NaN   \n",
       "1                                 NaN                        NaN       NaN   \n",
       "2                                 NaN                        NaN       NaN   \n",
       "3                                 NaN  chr16:076331867-076331926  hs|q23.1   \n",
       "4         ref|NM_138295|ref|NM_025031   chr7:047626734-047626793  hs|p12.3   \n",
       "...                               ...                        ...       ...   \n",
       "15092790                          NaN                        NaN       NaN   \n",
       "15092791                          NaN                        NaN       NaN   \n",
       "15092792                          NaN                        NaN       NaN   \n",
       "15092793                          NaN                        NaN       NaN   \n",
       "15092794                          NaN                        NaN       NaN   \n",
       "\n",
       "                                                DESCRIPTION  \\\n",
       "0                                                       NaN   \n",
       "1                                                       NaN   \n",
       "2                                                       NaN   \n",
       "3                                                       NaN   \n",
       "4         Homo sapiens polycystic kidney disease 1 like ...   \n",
       "...                                                     ...   \n",
       "15092790                                                NaN   \n",
       "15092791                                                NaN   \n",
       "15092792                                                NaN   \n",
       "15092793                                                NaN   \n",
       "15092794                                                NaN   \n",
       "\n",
       "                                    GB_RANGE  \n",
       "0                                        NaN  \n",
       "1                                        NaN  \n",
       "2                                        NaN  \n",
       "3          NC_000016.8[076331867..076331926]  \n",
       "4         NC_000007.11[047626734..047626793]  \n",
       "...                                      ...  \n",
       "15092790                                 NaN  \n",
       "15092791                                 NaN  \n",
       "15092792                                 NaN  \n",
       "15092793                                 NaN  \n",
       "15092794                                 NaN  \n",
       "\n",
       "[15092795 rows x 13 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, nan, nan, nan, 'PKD1L1']"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation_summary['GENE_SYMBOL']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GENE_SYMBOL'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '13', '14', '15', '16', '17', '18', '19', \n",
      "1000 input query terms found no hit:\t['1002', '1003', '1004', '1005', '1006', '1007', '1008', '1009', '1010', '1011', '1012', '1013', '10\n",
      "1000 input query terms found no hit:\t['2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '20\n",
      "1000 input query terms found no hit:\t['3005', '3006', '3007', '3008', '3009', '3010', '3011', '3012', '3013', '3014', '3015', '3016', '30\n",
      "1000 input query terms found no hit:\t['4005', '4006', '4007', '4008', '4009', '4010', '4011', '4012', '4013', '4014', '4015', '4016', '40\n",
      "1000 input query terms found no hit:\t['5005', '5006', '5007', '5008', '5009', '5010', '5011', '5012', '5013', '5014', '5015', '5016', '50\n",
      "1000 input query terms found no hit:\t['6005', '6006', '6007', '6008', '6009', '6010', '6011', '6012', '6013', '6014', '6015', '6016', '60\n",
      "1000 input query terms found no hit:\t['7006', '7007', '7008', '7009', '7010', '7011', '7012', '7013', '7014', '7015', '7016', '7017', '70\n",
      "1000 input query terms found no hit:\t['8006', '8007', '8008', '8009', '8010', '8011', '8012', '8013', '8014', '8015', '8016', '8017', '80\n",
      "1000 input query terms found no hit:\t['9007', '9008', '9009', '9010', '9011', '9012', '9013', '9014', '9015', '9016', '9017', '9018', '90\n",
      "1000 input query terms found no hit:\t['10007', '10008', '10009', '10010', '10011', '10012', '10013', '10014', '10015', '10016', '10017', \n",
      "1000 input query terms found no hit:\t['11007', '11008', '11009', '11010', '11011', '11012', '11013', '11014', '11015', '11016', '11017', \n",
      "1000 input query terms found no hit:\t['12008', '12009', '12010', '12011', '12012', '12013', '12014', '12015', '12016', '12017', '12018', \n",
      "1000 input query terms found no hit:\t['13008', '13009', '13010', '13011', '13012', '13013', '13014', '13015', '13016', '13017', '13018', \n",
      "1000 input query terms found no hit:\t['14009', '14010', '14011', '14012', '14013', '14014', '14015', '14016', '14017', '14018', '14019', \n",
      "1000 input query terms found no hit:\t['15010', '15011', '15012', '15013', '15014', '15015', '15016', '15017', '15018', '15019', '15020', \n",
      "1000 input query terms found no hit:\t['16011', '16012', '16013', '16014', '16015', '16016', '16017', '16018', '16019', '16020', '16021', \n",
      "1000 input query terms found no hit:\t['17011', '17012', '17013', '17014', '17015', '17016', '17017', '17018', '17019', '17020', '17021', \n",
      "1000 input query terms found no hit:\t['18011', '18012', '18013', '18014', '18015', '18016', '18017', '18018', '18019', '18020', '18021', \n",
      "1000 input query terms found no hit:\t['19011', '19012', '19013', '19014', '19015', '19016', '19017', '19018', '19019', '19020', '19021', \n",
      "1000 input query terms found no hit:\t['20011', '20012', '20013', '20014', '20015', '20016', '20017', '20018', '20019', '20020', '20021', \n",
      "1000 input query terms found no hit:\t['21012', '21013', '21014', '21015', '21016', '21017', '21018', '21019', '21020', '21021', '21022', \n",
      "1000 input query terms found no hit:\t['22013', '22014', '22015', '22016', '22017', '22018', '22019', '22020', '22021', '22022', '22023', \n",
      "1000 input query terms found no hit:\t['23013', '23014', '23015', '23016', '23017', '23018', '23019', '23020', '23021', '23022', '23023', \n",
      "1000 input query terms found no hit:\t['24013', '24014', '24015', '24016', '24017', '24018', '24019', '24020', '24021', '24022', '24023', \n",
      "1000 input query terms found no hit:\t['25013', '25014', '25015', '25016', '25017', '25018', '25019', '25020', '25021', '25022', '25023', \n",
      "1000 input query terms found no hit:\t['26013', '26014', '26015', '26016', '26017', '26018', '26019', '26020', '26021', '26022', '26023', \n",
      "1000 input query terms found no hit:\t['27013', '27014', '27015', '27016', '27017', '27018', '27019', '27020', '27021', '27022', '27023', \n",
      "1000 input query terms found no hit:\t['28013', '28014', '28015', '28016', '28017', '28018', '28019', '28020', '28021', '28022', '28023', \n",
      "1000 input query terms found no hit:\t['29015', '29016', '29017', '29018', '29019', '29020', '29021', '29022', '29023', '29024', '29025', \n",
      "1000 input query terms found no hit:\t['30015', '30016', '30017', '30018', '30019', '30020', '30021', '30022', '30023', '30024', '30025', \n",
      "1000 input query terms found no hit:\t['31015', '31016', '31017', '31018', '31019', '31020', '31021', '31022', '31023', '31024', '31025', \n",
      "1000 input query terms found no hit:\t['32015', '32016', '32017', '32018', '32019', '32020', '32021', '32022', '32023', '32024', '32025', \n",
      "1000 input query terms found no hit:\t['33015', '33016', '33017', '33018', '33019', '33020', '33021', '33022', '33023', '33024', '33025', \n",
      "1000 input query terms found no hit:\t['34015', '34016', '34017', '34018', '34019', '34020', '34021', '34022', '34023', '34024', '34025', \n",
      "1000 input query terms found no hit:\t['35015', '35016', '35017', '35018', '35019', '35020', '35021', '35022', '35023', '35024', '35025', \n",
      "1000 input query terms found no hit:\t['36015', '36016', '36017', '36018', '36019', '36020', '36021', '36022', '36023', '36024', '36025', \n",
      "1000 input query terms found no hit:\t['37016', '37017', '37018', '37019', '37020', '37021', '37022', '37023', '37024', '37025', '37026', \n",
      "1000 input query terms found no hit:\t['38016', '38017', '38018', '38019', '38020', '38021', '38022', '38023', '38024', '38025', '38026', \n",
      "1000 input query terms found no hit:\t['39017', '39018', '39019', '39020', '39021', '39022', '39023', '39024', '39025', '39026', '39027', \n",
      "1000 input query terms found no hit:\t['40017', '40018', '40019', '40020', '40021', '40022', '40023', '40024', '40025', '40026', '40027', \n",
      "1000 input query terms found no hit:\t['41017', '41018', '41019', '41020', '41021', '41022', '41023', '41024', '41025', '41026', '41027', \n",
      "1000 input query terms found no hit:\t['42017', '42018', '42019', '42020', '42021', '42022', '42023', '42024', '42025', '42026', '42027', \n",
      "1000 input query terms found no hit:\t['43017', '43018', '43019', '43020', '43021', '43022', '43023', '43024', '43025', '43026', '43027', \n",
      "1000 input query terms found no hit:\t['44017', '44018', '44019', '44020', '44021', '44022', '44023', '44024', '44025', '44026', '44027', \n",
      "1000 input query terms found no hit:\t['45017', '45018', '45019', '45020', '45021', '45022', '45023', '45024', '45025', '45026', '45027', \n",
      "1000 input query terms found no hit:\t['46017', '46018', '46019', '46020', '46021', '46022', '46023', '46024', '46025', '46026', '46027', \n",
      "1000 input query terms found no hit:\t['47017', '47018', '47019', '47020', '47021', '47022', '47023', '47024', '47025', '47026', '47027', \n",
      "1000 input query terms found no hit:\t['48017', '48018', '48019', '48020', '48021', '48022', '48023', '48024', '48025', '48026', '48027', \n",
      "1000 input query terms found no hit:\t['49017', '49018', '49019', '49020', '49021', '49022', '49023', '49024', '49025', '49026', '49027', \n",
      "1000 input query terms found no hit:\t['50017', '50018', '50019', '50020', '50021', '50022', '50023', '50024', '50025', '50026', '50027', \n",
      "1000 input query terms found no hit:\t['51017', '51018', '51019', '51020', '51021', '51022', '51023', '51024', '51025', '51026', '51027', \n",
      "1000 input query terms found no hit:\t['52017', '52018', '52019', '52020', '52021', '52022', '52023', '52024', '52025', '52026', '52027', \n",
      "1000 input query terms found no hit:\t['53017', '53018', '53019', '53020', '53021', '53022', '53023', '53024', '53025', '53026', '53027', \n",
      "1000 input query terms found no hit:\t['54018', '54019', '54020', '54021', '54022', '54023', '54024', '54025', '54026', '54027', '54028', \n",
      "1000 input query terms found no hit:\t['55018', '55019', '55020', '55021', '55022', '55023', '55024', '55025', '55026', '55027', '55028', \n",
      "1000 input query terms found no hit:\t['56018', '56019', '56020', '56021', '56022', '56023', '56024', '56025', '56026', '56027', '56028', \n",
      "1000 input query terms found no hit:\t['57019', '57020', '57021', '57022', '57023', '57024', '57025', '57026', '57027', '57028', '57029', \n",
      "1000 input query terms found no hit:\t['58019', '58020', '58021', '58022', '58023', '58024', '58025', '58026', '58027', '58028', '58029', \n",
      "1000 input query terms found no hit:\t['59019', '59020', '59021', '59022', '59023', '59024', '59025', '59026', '59027', '59028', '59029', \n",
      "1000 input query terms found no hit:\t['60019', '60020', '60021', '60022', '60023', '60024', '60025', '60026', '60027', '60028', '60029', \n",
      "1000 input query terms found no hit:\t['61019', '61020', '61021', '61022', '61023', '61024', '61025', '61026', '61027', '61028', '61029', \n",
      "1000 input query terms found no hit:\t['62019', '62020', '62021', '62022', '62023', '62024', '62025', '62026', '62027', '62028', '62029', \n",
      "1000 input query terms found no hit:\t['63019', '63020', '63021', '63022', '63023', '63024', '63025', '63026', '63027', '63028', '63029', \n",
      "1000 input query terms found no hit:\t['64019', '64020', '64021', '64022', '64023', '64024', '64025', '64026', '64027', '64028', '64029', \n",
      "1000 input query terms found no hit:\t['65019', '65020', '65021', '65022', '65023', '65024', '65025', '65026', '65027', '65028', '65029', \n",
      "1000 input query terms found no hit:\t['66019', '66020', '66021', '66022', '66023', '66024', '66025', '66026', '66027', '66028', '66029', \n",
      "1000 input query terms found no hit:\t['67019', '67020', '67021', '67022', '67023', '67024', '67025', '67026', '67027', '67028', '67029', \n",
      "1000 input query terms found no hit:\t['68020', '68021', '68022', '68023', '68024', '68025', '68026', '68027', '68028', '68029', '68030', \n",
      "1000 input query terms found no hit:\t['69020', '69021', '69022', '69023', '69024', '69025', '69026', '69027', '69028', '69029', '69030', \n",
      "1000 input query terms found no hit:\t['70021', '70022', '70023', '70024', '70025', '70026', '70027', '70028', '70029', '70030', '70031', \n",
      "1000 input query terms found no hit:\t['71021', '71022', '71023', '71024', '71025', '71026', '71027', '71028', '71029', '71030', '71031', \n",
      "1000 input query terms found no hit:\t['72022', '72023', '72024', '72025', '72026', '72027', '72028', '72029', '72030', '72031', '72032', \n",
      "1000 input query terms found no hit:\t['73022', '73023', '73024', '73025', '73026', '73027', '73028', '73029', '73030', '73031', '73032', \n",
      "1000 input query terms found no hit:\t['74022', '74023', '74024', '74025', '74026', '74027', '74028', '74029', '74030', '74031', '74032', \n",
      "1000 input query terms found no hit:\t['75022', '75023', '75024', '75025', '75026', '75027', '75028', '75029', '75030', '75031', '75032', \n",
      "1000 input query terms found no hit:\t['76023', '76024', '76025', '76026', '76027', '76028', '76029', '76030', '76031', '76032', '76033', \n",
      "1000 input query terms found no hit:\t['77023', '77024', '77025', '77026', '77027', '77028', '77029', '77030', '77031', '77032', '77033', \n",
      "1000 input query terms found no hit:\t['78023', '78024', '78025', '78026', '78027', '78028', '78029', '78030', '78031', '78032', '78033', \n",
      "1000 input query terms found no hit:\t['79023', '79024', '79025', '79026', '79027', '79028', '79029', '79030', '79031', '79032', '79033', \n",
      "1000 input query terms found no hit:\t['80023', '80024', '80025', '80026', '80027', '80028', '80029', '80030', '80031', '80032', '80033', \n",
      "1000 input query terms found no hit:\t['81023', '81024', '81025', '81026', '81027', '81028', '81029', '81030', '81031', '81032', '81033', \n",
      "1000 input query terms found no hit:\t['82023', '82024', '82025', '82026', '82027', '82028', '82029', '82030', '82031', '82032', '82033', \n",
      "1000 input query terms found no hit:\t['83024', '83025', '83026', '83027', '83028', '83029', '83030', '83031', '83032', '83033', '83034', \n",
      "1000 input query terms found no hit:\t['84024', '84025', '84026', '84027', '84028', '84029', '84030', '84031', '84032', '84033', '84034', \n",
      "1000 input query terms found no hit:\t['85025', '85026', '85027', '85028', '85029', '85030', '85031', '85032', '85033', '85034', '85035', \n",
      "1000 input query terms found no hit:\t['86025', '86026', '86027', '86028', '86029', '86030', '86031', '86032', '86033', '86034', '86035', \n",
      "1000 input query terms found no hit:\t['87025', '87026', '87027', '87028', '87029', '87030', '87031', '87032', '87033', '87034', '87035', \n",
      "1000 input query terms found no hit:\t['88025', '88026', '88027', '88028', '88029', '88030', '88031', '88032', '88033', '88034', '88035', \n",
      "1000 input query terms found no hit:\t['89025', '89026', '89027', '89028', '89029', '89030', '89031', '89032', '89033', '89034', '89035', \n",
      "1000 input query terms found no hit:\t['90025', '90026', '90027', '90028', '90029', '90030', '90031', '90032', '90033', '90034', '90035', \n",
      "1000 input query terms found no hit:\t['91026', '91027', '91028', '91029', '91030', '91031', '91032', '91033', '91034', '91035', '91036', \n",
      "1000 input query terms found no hit:\t['92027', '92028', '92029', '92030', '92031', '92032', '92033', '92034', '92035', '92036', '92037', \n",
      "1000 input query terms found no hit:\t['93027', '93028', '93029', '93030', '93031', '93032', '93033', '93034', '93035', '93036', '93037', \n",
      "1000 input query terms found no hit:\t['94027', '94028', '94029', '94030', '94031', '94032', '94033', '94034', '94035', '94036', '94037', \n",
      "1000 input query terms found no hit:\t['95027', '95028', '95029', '95030', '95031', '95032', '95033', '95034', '95035', '95036', '95037', \n",
      "1000 input query terms found no hit:\t['96027', '96028', '96029', '96030', '96031', '96032', '96033', '96034', '96035', '96036', '96037', \n",
      "1000 input query terms found no hit:\t['97027', '97028', '97029', '97030', '97031', '97032', '97033', '97034', '97035', '97036', '97037', \n",
      "1000 input query terms found no hit:\t['98027', '98028', '98029', '98030', '98031', '98032', '98033', '98034', '98035', '98036', '98037', \n",
      "1000 input query terms found no hit:\t['99027', '99028', '99029', '99030', '99031', '99032', '99033', '99034', '99035', '99036', '99037', \n",
      "1000 input query terms found no hit:\t['100027', '100028', '100029', '100030', '100031', '100032', '100033', '100034', '100035', '100036',\n",
      "1000 input query terms found no hit:\t['101027', '101028', '101029', '101030', '101031', '101032', '101033', '101034', '101035', '101036',\n",
      "1000 input query terms found no hit:\t['102027', '102028', '102029', '102030', '102031', '102032', '102033', '102034', '102035', '102036',\n",
      "1000 input query terms found no hit:\t['103027', '103028', '103029', '103030', '103031', '103032', '103033', '103034', '103035', '103036',\n",
      "1000 input query terms found no hit:\t['104027', '104028', '104029', '104030', '104031', '104032', '104033', '104034', '104035', '104036',\n",
      "1000 input query terms found no hit:\t['105027', '105028', '105029', '105030', '105031', '105032', '105033', '105034', '105035', '105036',\n",
      "1000 input query terms found no hit:\t['106027', '106028', '106029', '106030', '106031', '106032', '106033', '106034', '106035', '106036',\n",
      "1000 input query terms found no hit:\t['107028', '107029', '107030', '107031', '107032', '107033', '107034', '107035', '107036', '107037',\n",
      "1000 input query terms found no hit:\t['108028', '108029', '108030', '108031', '108032', '108033', '108034', '108035', '108036', '108037',\n",
      "1000 input query terms found no hit:\t['109028', '109029', '109030', '109031', '109032', '109033', '109034', '109035', '109036', '109037',\n",
      "1000 input query terms found no hit:\t['110028', '110029', '110030', '110031', '110032', '110033', '110034', '110035', '110036', '110037',\n",
      "1000 input query terms found no hit:\t['111029', '111030', '111031', '111032', '111033', '111034', '111035', '111036', '111037', '111038',\n",
      "1000 input query terms found no hit:\t['112029', '112030', '112031', '112032', '112033', '112034', '112035', '112036', '112037', '112038',\n",
      "1000 input query terms found no hit:\t['113029', '113030', '113031', '113032', '113033', '113034', '113035', '113036', '113037', '113038',\n",
      "1000 input query terms found no hit:\t['114031', '114032', '114033', '114034', '114035', '114036', '114037', '114038', '114039', '114040',\n",
      "1000 input query terms found no hit:\t['115031', '115032', '115033', '115034', '115035', '115036', '115037', '115038', '115039', '115040',\n",
      "1000 input query terms found no hit:\t['116031', '116032', '116033', '116034', '116035', '116036', '116037', '116038', '116039', '116040',\n",
      "1000 input query terms found no hit:\t['117031', '117032', '117033', '117034', '117035', '117036', '117037', '117038', '117039', '117040',\n",
      "1000 input query terms found no hit:\t['118031', '118032', '118033', '118034', '118035', '118036', '118037', '118038', '118039', '118040',\n",
      "1000 input query terms found no hit:\t['119032', '119033', '119034', '119035', '119036', '119037', '119038', '119039', '119040', '119041',\n",
      "1000 input query terms found no hit:\t['120032', '120033', '120034', '120035', '120036', '120037', '120038', '120039', '120040', '120041',\n",
      "1000 input query terms found no hit:\t['121032', '121033', '121034', '121035', '121036', '121037', '121038', '121039', '121040', '121041',\n",
      "1000 input query terms found no hit:\t['122033', '122034', '122035', '122036', '122037', '122038', '122039', '122040', '122041', '122042',\n",
      "1000 input query terms found no hit:\t['123033', '123034', '123035', '123036', '123037', '123038', '123039', '123040', '123041', '123042',\n",
      "1000 input query terms found no hit:\t['124034', '124035', '124036', '124037', '124038', '124039', '124040', '124041', '124042', '124043',\n",
      "1000 input query terms found no hit:\t['125036', '125037', '125038', '125039', '125040', '125041', '125042', '125043', '125044', '125045',\n",
      "1000 input query terms found no hit:\t['126037', '126038', '126039', '126040', '126041', '126042', '126043', '126044', '126045', '126046',\n",
      "1000 input query terms found no hit:\t['127037', '127038', '127039', '127040', '127041', '127042', '127043', '127044', '127045', '127046',\n",
      "1000 input query terms found no hit:\t['128037', '128038', '128039', '128040', '128041', '128042', '128043', '128044', '128045', '128046',\n",
      "1000 input query terms found no hit:\t['129037', '129038', '129039', '129040', '129041', '129042', '129043', '129044', '129045', '129046',\n",
      "1000 input query terms found no hit:\t['130037', '130038', '130039', '130040', '130041', '130042', '130043', '130044', '130045', '130046',\n",
      "1000 input query terms found no hit:\t['131038', '131039', '131040', '131041', '131042', '131043', '131044', '131045', '131046', '131047',\n",
      "1000 input query terms found no hit:\t['132038', '132039', '132040', '132041', '132042', '132043', '132044', '132045', '132046', '132047',\n",
      "1000 input query terms found no hit:\t['133038', '133039', '133040', '133041', '133042', '133043', '133044', '133045', '133046', '133047',\n",
      "1000 input query terms found no hit:\t['134040', '134041', '134042', '134043', '134044', '134045', '134046', '134047', '134048', '134049',\n",
      "1000 input query terms found no hit:\t['135040', '135041', '135042', '135043', '135044', '135045', '135046', '135047', '135048', '135049',\n",
      "1000 input query terms found no hit:\t['136040', '136041', '136042', '136043', '136044', '136045', '136046', '136047', '136048', '136049',\n",
      "1000 input query terms found no hit:\t['137040', '137041', '137042', '137043', '137044', '137045', '137046', '137047', '137048', '137049',\n",
      "1000 input query terms found no hit:\t['138040', '138041', '138042', '138043', '138044', '138045', '138046', '138047', '138048', '138049',\n",
      "1000 input query terms found no hit:\t['139040', '139041', '139042', '139043', '139044', '139045', '139046', '139047', '139048', '139049',\n",
      "1000 input query terms found no hit:\t['140040', '140041', '140042', '140043', '140044', '140045', '140046', '140047', '140048', '140049',\n",
      "1000 input query terms found no hit:\t['141041', '141042', '141043', '141044', '141045', '141046', '141047', '141048', '141049', '141050',\n",
      "1000 input query terms found no hit:\t['142042', '142043', '142044', '142045', '142046', '142047', '142048', '142049', '142050', '142051',\n",
      "1000 input query terms found no hit:\t['143042', '143043', '143044', '143045', '143046', '143047', '143048', '143049', '143050', '143051',\n",
      "1000 input query terms found no hit:\t['144043', '144044', '144045', '144046', '144047', '144048', '144049', '144050', '144051', '144052',\n",
      "1000 input query terms found no hit:\t['145043', '145044', '145045', '145046', '145047', '145048', '145049', '145050', '145051', '145052',\n",
      "1000 input query terms found no hit:\t['146043', '146044', '146045', '146046', '146047', '146048', '146049', '146050', '146051', '146052',\n",
      "1000 input query terms found no hit:\t['147044', '147045', '147046', '147047', '147048', '147049', '147050', '147051', '147052', '147053',\n",
      "1000 input query terms found no hit:\t['148044', '148045', '148046', '148047', '148048', '148049', '148050', '148051', '148052', '148053',\n",
      "1000 input query terms found no hit:\t['149046', '149047', '149048', '149049', '149050', '149051', '149052', '149053', '149054', '149055',\n",
      "1000 input query terms found no hit:\t['150046', '150047', '150048', '150049', '150050', '150051', '150052', '150053', '150054', '150055',\n",
      "1000 input query terms found no hit:\t['151046', '151047', '151048', '151049', '151050', '151051', '151052', '151053', '151054', '151055',\n",
      "1000 input query terms found no hit:\t['152047', '152048', '152049', '152050', '152051', '152052', '152053', '152054', '152055', '152056',\n",
      "1000 input query terms found no hit:\t['153048', '153049', '153050', '153051', '153052', '153053', '153054', '153055', '153056', '153057',\n",
      "1000 input query terms found no hit:\t['154048', '154049', '154050', '154051', '154052', '154053', '154054', '154055', '154056', '154057',\n",
      "1000 input query terms found no hit:\t['155048', '155049', '155050', '155051', '155052', '155053', '155054', '155055', '155056', '155057',\n",
      "1000 input query terms found no hit:\t['156048', '156049', '156050', '156051', '156052', '156053', '156054', '156055', '156056', '156057',\n",
      "1000 input query terms found no hit:\t['157048', '157049', '157050', '157051', '157052', '157053', '157054', '157055', '157056', '157057',\n",
      "1000 input query terms found no hit:\t['158048', '158049', '158050', '158051', '158052', '158053', '158054', '158055', '158056', '158057',\n",
      "1000 input query terms found no hit:\t['159049', '159050', '159051', '159052', '159053', '159054', '159055', '159056', '159057', '159058',\n",
      "1000 input query terms found no hit:\t['160050', '160051', '160052', '160053', '160054', '160055', '160056', '160057', '160058', '160059',\n",
      "1000 input query terms found no hit:\t['161050', '161051', '161052', '161053', '161054', '161055', '161056', '161057', '161058', '161059',\n",
      "1000 input query terms found no hit:\t['162050', '162051', '162052', '162053', '162054', '162055', '162056', '162057', '162058', '162059',\n",
      "1000 input query terms found no hit:\t['163050', '163051', '163052', '163053', '163054', '163055', '163056', '163057', '163058', '163059',\n",
      "1000 input query terms found no hit:\t['164050', '164051', '164052', '164053', '164054', '164055', '164056', '164057', '164058', '164059',\n",
      "1000 input query terms found no hit:\t['165050', '165051', '165052', '165053', '165054', '165055', '165056', '165057', '165058', '165059',\n",
      "1000 input query terms found no hit:\t['166052', '166053', '166054', '166055', '166056', '166057', '166058', '166059', '166060', '166061',\n",
      "1000 input query terms found no hit:\t['167052', '167053', '167054', '167055', '167056', '167057', '167058', '167059', '167060', '167061',\n",
      "1000 input query terms found no hit:\t['168053', '168054', '168055', '168056', '168057', '168058', '168059', '168060', '168061', '168062',\n",
      "1000 input query terms found no hit:\t['169054', '169055', '169056', '169057', '169058', '169059', '169060', '169061', '169062', '169063',\n",
      "1000 input query terms found no hit:\t['170054', '170055', '170056', '170057', '170058', '170059', '170060', '170061', '170062', '170063',\n",
      "1000 input query terms found no hit:\t['171054', '171055', '171056', '171057', '171058', '171059', '171060', '171061', '171062', '171063',\n",
      "1000 input query terms found no hit:\t['172055', '172056', '172057', '172058', '172059', '172060', '172061', '172062', '172063', '172064',\n",
      "1000 input query terms found no hit:\t['173055', '173056', '173057', '173058', '173059', '173060', '173061', '173062', '173063', '173064',\n",
      "1000 input query terms found no hit:\t['174055', '174056', '174057', '174058', '174059', '174060', '174061', '174062', '174063', '174064',\n",
      "1000 input query terms found no hit:\t['175056', '175057', '175058', '175059', '175060', '175061', '175062', '175063', '175064', '175065',\n",
      "1000 input query terms found no hit:\t['176056', '176057', '176058', '176059', '176060', '176061', '176062', '176063', '176064', '176065',\n",
      "1000 input query terms found no hit:\t['177056', '177057', '177058', '177059', '177060', '177061', '177062', '177063', '177064', '177065',\n",
      "1000 input query terms found no hit:\t['178056', '178057', '178058', '178059', '178060', '178061', '178062', '178063', '178064', '178065',\n",
      "1000 input query terms found no hit:\t['179056', '179057', '179058', '179059', '179060', '179061', '179062', '179063', '179064', '179065',\n",
      "1000 input query terms found no hit:\t['180056', '180057', '180058', '180059', '180060', '180061', '180062', '180063', '180064', '180065',\n",
      "1000 input query terms found no hit:\t['181056', '181057', '181058', '181059', '181060', '181061', '181062', '181063', '181064', '181065',\n",
      "1000 input query terms found no hit:\t['182057', '182058', '182059', '182060', '182061', '182062', '182063', '182064', '182065', '182066',\n",
      "1000 input query terms found no hit:\t['183057', '183058', '183059', '183060', '183061', '183062', '183063', '183064', '183065', '183066',\n",
      "1000 input query terms found no hit:\t['184057', '184058', '184059', '184060', '184061', '184062', '184063', '184064', '184065', '184066',\n",
      "1000 input query terms found no hit:\t['185057', '185058', '185059', '185060', '185061', '185062', '185063', '185064', '185065', '185066',\n",
      "1000 input query terms found no hit:\t['186057', '186058', '186059', '186060', '186061', '186062', '186063', '186064', '186065', '186066',\n",
      "1000 input query terms found no hit:\t['187057', '187058', '187059', '187060', '187061', '187062', '187063', '187064', '187065', '187066',\n",
      "1000 input query terms found no hit:\t['188058', '188059', '188060', '188061', '188062', '188063', '188064', '188065', '188066', '188067',\n",
      "1000 input query terms found no hit:\t['189058', '189059', '189060', '189061', '189062', '189063', '189064', '189065', '189066', '189067',\n",
      "1000 input query terms found no hit:\t['190058', '190059', '190060', '190061', '190062', '190063', '190064', '190065', '190066', '190067',\n",
      "1000 input query terms found no hit:\t['191060', '191061', '191062', '191063', '191064', '191065', '191066', '191067', '191068', '191069',\n",
      "1000 input query terms found no hit:\t['192060', '192061', '192062', '192063', '192064', '192065', '192066', '192067', '192068', '192069',\n",
      "1000 input query terms found no hit:\t['193060', '193061', '193062', '193063', '193064', '193065', '193066', '193067', '193068', '193069',\n",
      "1000 input query terms found no hit:\t['194060', '194061', '194062', '194063', '194064', '194065', '194066', '194067', '194068', '194069',\n",
      "1000 input query terms found no hit:\t['195061', '195062', '195063', '195064', '195065', '195066', '195067', '195068', '195069', '195070',\n",
      "1000 input query terms found no hit:\t['196061', '196062', '196063', '196064', '196065', '196066', '196067', '196068', '196069', '196070',\n",
      "1000 input query terms found no hit:\t['197062', '197063', '197064', '197065', '197066', '197067', '197068', '197069', '197070', '197071',\n",
      "1000 input query terms found no hit:\t['198063', '198064', '198065', '198066', '198067', '198068', '198069', '198070', '198071', '198072',\n",
      "1000 input query terms found no hit:\t['199063', '199064', '199065', '199066', '199067', '199068', '199069', '199070', '199071', '199072',\n",
      "1000 input query terms found no hit:\t['200063', '200064', '200065', '200066', '200067', '200068', '200069', '200070', '200071', '200072',\n",
      "1000 input query terms found no hit:\t['201064', '201065', '201066', '201067', '201068', '201069', '201070', '201071', '201072', '201073',\n",
      "1000 input query terms found no hit:\t['202064', '202065', '202066', '202067', '202068', '202069', '202070', '202071', '202072', '202073',\n",
      "1000 input query terms found no hit:\t['203065', '203066', '203067', '203068', '203069', '203070', '203071', '203072', '203073', '203074',\n",
      "1000 input query terms found no hit:\t['204065', '204066', '204067', '204068', '204069', '204070', '204071', '204072', '204073', '204074',\n",
      "1000 input query terms found no hit:\t['205066', '205067', '205068', '205069', '205070', '205071', '205072', '205073', '205074', '205075',\n",
      "1000 input query terms found no hit:\t['206066', '206067', '206068', '206069', '206070', '206071', '206072', '206073', '206074', '206075',\n",
      "1000 input query terms found no hit:\t['207066', '207067', '207068', '207069', '207070', '207071', '207072', '207073', '207074', '207075',\n",
      "1000 input query terms found no hit:\t['208066', '208067', '208068', '208069', '208070', '208071', '208072', '208073', '208074', '208075',\n",
      "1000 input query terms found no hit:\t['209066', '209067', '209068', '209069', '209070', '209071', '209072', '209073', '209074', '209075',\n",
      "1000 input query terms found no hit:\t['210066', '210067', '210068', '210069', '210070', '210071', '210072', '210073', '210074', '210075',\n",
      "1000 input query terms found no hit:\t['211066', '211067', '211068', '211069', '211070', '211071', '211072', '211073', '211074', '211075',\n",
      "1000 input query terms found no hit:\t['212067', '212068', '212069', '212070', '212071', '212072', '212073', '212074', '212075', '212076',\n",
      "1000 input query terms found no hit:\t['213068', '213069', '213070', '213071', '213072', '213073', '213074', '213075', '213076', '213077',\n",
      "1000 input query terms found no hit:\t['214068', '214069', '214070', '214071', '214072', '214073', '214074', '214075', '214076', '214077',\n",
      "1000 input query terms found no hit:\t['215068', '215069', '215070', '215071', '215072', '215073', '215074', '215075', '215076', '215077',\n",
      "1000 input query terms found no hit:\t['216068', '216069', '216070', '216071', '216072', '216073', '216074', '216075', '216076', '216077',\n",
      "1000 input query terms found no hit:\t['217068', '217069', '217070', '217071', '217072', '217073', '217074', '217075', '217076', '217077',\n",
      "1000 input query terms found no hit:\t['218068', '218069', '218070', '218071', '218072', '218073', '218074', '218075', '218076', '218077',\n",
      "1000 input query terms found no hit:\t['219069', '219070', '219071', '219072', '219073', '219074', '219075', '219076', '219077', '219078',\n",
      "1000 input query terms found no hit:\t['220070', '220071', '220072', '220073', '220074', '220075', '220076', '220077', '220078', '220079',\n",
      "1000 input query terms found no hit:\t['221071', '221072', '221073', '221074', '221075', '221076', '221077', '221078', '221079', '221080',\n",
      "1000 input query terms found no hit:\t['222071', '222072', '222073', '222074', '222075', '222076', '222077', '222078', '222079', '222080',\n",
      "1000 input query terms found no hit:\t['223071', '223072', '223073', '223074', '223075', '223076', '223077', '223078', '223079', '223080',\n",
      "1000 input query terms found no hit:\t['224071', '224072', '224073', '224074', '224075', '224076', '224077', '224078', '224079', '224080',\n",
      "1000 input query terms found no hit:\t['225071', '225072', '225073', '225074', '225075', '225076', '225077', '225078', '225079', '225080',\n",
      "1000 input query terms found no hit:\t['226071', '226072', '226073', '226074', '226075', '226076', '226077', '226078', '226079', '226080',\n",
      "1000 input query terms found no hit:\t['227071', '227072', '227073', '227074', '227075', '227076', '227077', '227078', '227079', '227080',\n",
      "1000 input query terms found no hit:\t['228071', '228072', '228073', '228074', '228075', '228076', '228077', '228078', '228079', '228080',\n",
      "1000 input query terms found no hit:\t['229071', '229072', '229073', '229074', '229075', '229076', '229077', '229078', '229079', '229080',\n",
      "1000 input query terms found no hit:\t['230072', '230073', '230074', '230075', '230076', '230077', '230078', '230079', '230080', '230081',\n",
      "1000 input query terms found no hit:\t['231073', '231074', '231075', '231076', '231077', '231078', '231079', '231080', '231081', '231082',\n",
      "1000 input query terms found no hit:\t['232074', '232075', '232076', '232077', '232078', '232079', '232080', '232081', '232082', '232083',\n",
      "1000 input query terms found no hit:\t['233074', '233075', '233076', '233077', '233078', '233079', '233080', '233081', '233082', '233083',\n",
      "1000 input query terms found no hit:\t['234075', '234076', '234077', '234078', '234079', '234080', '234081', '234082', '234083', '234084',\n",
      "1000 input query terms found no hit:\t['235075', '235076', '235077', '235078', '235079', '235080', '235081', '235082', '235083', '235084',\n",
      "1000 input query terms found no hit:\t['236076', '236077', '236078', '236079', '236080', '236081', '236082', '236083', '236084', '236085',\n",
      "1000 input query terms found no hit:\t['237076', '237077', '237078', '237079', '237080', '237081', '237082', '237083', '237084', '237085',\n",
      "1000 input query terms found no hit:\t['238076', '238077', '238078', '238079', '238080', '238081', '238082', '238083', '238084', '238085',\n",
      "1000 input query terms found no hit:\t['239076', '239077', '239078', '239079', '239080', '239081', '239082', '239083', '239084', '239085',\n",
      "1000 input query terms found no hit:\t['240076', '240077', '240078', '240079', '240080', '240081', '240082', '240083', '240084', '240085',\n",
      "1000 input query terms found no hit:\t['241076', '241077', '241078', '241079', '241080', '241081', '241082', '241083', '241084', '241085',\n",
      "1000 input query terms found no hit:\t['242076', '242077', '242078', '242079', '242080', '242081', '242082', '242083', '242084', '242085',\n",
      "429 input query terms found no hit:\t['243076', '243077', '243078', '243079', '243080', '243081', '243082', '243083', '243084', '243085',\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th></th>\n",
       "      <th>15E1.2</th>\n",
       "      <th>2'-PDE</th>\n",
       "      <th>76P</th>\n",
       "      <th>7A5</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2BP1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZYG11BL</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>bA16L21.2.1</th>\n",
       "      <th>bA9F11.1</th>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <th>hCAP-D3</th>\n",
       "      <th>hCAP-H2</th>\n",
       "      <th>mimitin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM540550</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.091366</td>\n",
       "      <td>-0.018767</td>\n",
       "      <td>-0.09565</td>\n",
       "      <td>-0.177225</td>\n",
       "      <td>-0.060111</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.110698</td>\n",
       "      <td>-0.071700</td>\n",
       "      <td>-0.056388</td>\n",
       "      <td>...</td>\n",
       "      <td>0.118914</td>\n",
       "      <td>-0.034700</td>\n",
       "      <td>-0.129433</td>\n",
       "      <td>-0.079166</td>\n",
       "      <td>0.029367</td>\n",
       "      <td>0.15015</td>\n",
       "      <td>0.00661</td>\n",
       "      <td>-0.104746</td>\n",
       "      <td>0.053800</td>\n",
       "      <td>-0.052605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540551</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.009954</td>\n",
       "      <td>-0.036233</td>\n",
       "      <td>-0.03820</td>\n",
       "      <td>-0.016975</td>\n",
       "      <td>0.072104</td>\n",
       "      <td>0.0237</td>\n",
       "      <td>0.088163</td>\n",
       "      <td>-0.075717</td>\n",
       "      <td>-0.025460</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128233</td>\n",
       "      <td>0.021967</td>\n",
       "      <td>-0.158727</td>\n",
       "      <td>-0.043763</td>\n",
       "      <td>0.102400</td>\n",
       "      <td>0.09530</td>\n",
       "      <td>-0.16100</td>\n",
       "      <td>-0.010522</td>\n",
       "      <td>0.027662</td>\n",
       "      <td>0.082836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540552</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.064026</td>\n",
       "      <td>0.126533</td>\n",
       "      <td>-0.27350</td>\n",
       "      <td>-0.015500</td>\n",
       "      <td>0.055131</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>-0.057415</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>0.121440</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077617</td>\n",
       "      <td>0.089867</td>\n",
       "      <td>-0.175639</td>\n",
       "      <td>-0.262375</td>\n",
       "      <td>-0.004233</td>\n",
       "      <td>-0.01800</td>\n",
       "      <td>-0.12000</td>\n",
       "      <td>0.026106</td>\n",
       "      <td>0.011612</td>\n",
       "      <td>0.032056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540553</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.223229</td>\n",
       "      <td>0.015967</td>\n",
       "      <td>-0.18850</td>\n",
       "      <td>-0.230250</td>\n",
       "      <td>-0.020300</td>\n",
       "      <td>0.1540</td>\n",
       "      <td>0.021842</td>\n",
       "      <td>0.176317</td>\n",
       "      <td>0.166813</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.034995</td>\n",
       "      <td>0.084667</td>\n",
       "      <td>-0.123181</td>\n",
       "      <td>-0.159617</td>\n",
       "      <td>-0.040867</td>\n",
       "      <td>0.26300</td>\n",
       "      <td>-0.13300</td>\n",
       "      <td>0.072607</td>\n",
       "      <td>0.063725</td>\n",
       "      <td>-0.053923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540554</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.015514</td>\n",
       "      <td>-0.001400</td>\n",
       "      <td>0.07720</td>\n",
       "      <td>-0.381500</td>\n",
       "      <td>0.201556</td>\n",
       "      <td>0.1410</td>\n",
       "      <td>0.034802</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>0.002068</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.074967</td>\n",
       "      <td>-0.098300</td>\n",
       "      <td>-0.021133</td>\n",
       "      <td>0.018503</td>\n",
       "      <td>-0.086867</td>\n",
       "      <td>0.03475</td>\n",
       "      <td>-0.13800</td>\n",
       "      <td>-0.014310</td>\n",
       "      <td>-0.114200</td>\n",
       "      <td>-0.005092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540606</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.135557</td>\n",
       "      <td>0.011360</td>\n",
       "      <td>0.01900</td>\n",
       "      <td>-0.010340</td>\n",
       "      <td>0.079336</td>\n",
       "      <td>-0.0340</td>\n",
       "      <td>0.066139</td>\n",
       "      <td>0.115183</td>\n",
       "      <td>-0.013030</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.131533</td>\n",
       "      <td>0.057200</td>\n",
       "      <td>-0.026909</td>\n",
       "      <td>-0.087575</td>\n",
       "      <td>0.167833</td>\n",
       "      <td>-0.15180</td>\n",
       "      <td>-0.12400</td>\n",
       "      <td>0.054733</td>\n",
       "      <td>-0.071883</td>\n",
       "      <td>-0.019704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540607</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125829</td>\n",
       "      <td>-0.109767</td>\n",
       "      <td>-0.11105</td>\n",
       "      <td>-0.110975</td>\n",
       "      <td>-0.004712</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>0.069542</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>-0.029762</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.153200</td>\n",
       "      <td>0.041467</td>\n",
       "      <td>-0.100020</td>\n",
       "      <td>-0.126593</td>\n",
       "      <td>0.107967</td>\n",
       "      <td>-0.28300</td>\n",
       "      <td>-0.27100</td>\n",
       "      <td>0.012011</td>\n",
       "      <td>-0.031462</td>\n",
       "      <td>0.012892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540608</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009514</td>\n",
       "      <td>0.300333</td>\n",
       "      <td>0.18200</td>\n",
       "      <td>0.131200</td>\n",
       "      <td>-0.077400</td>\n",
       "      <td>-0.1540</td>\n",
       "      <td>0.006996</td>\n",
       "      <td>0.168233</td>\n",
       "      <td>0.093887</td>\n",
       "      <td>...</td>\n",
       "      <td>0.372533</td>\n",
       "      <td>0.072733</td>\n",
       "      <td>0.083072</td>\n",
       "      <td>-0.300383</td>\n",
       "      <td>-0.159400</td>\n",
       "      <td>-0.07216</td>\n",
       "      <td>-0.17600</td>\n",
       "      <td>0.007011</td>\n",
       "      <td>0.068550</td>\n",
       "      <td>-0.295004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540609</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.174671</td>\n",
       "      <td>0.013167</td>\n",
       "      <td>-0.03220</td>\n",
       "      <td>-0.107950</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>0.2420</td>\n",
       "      <td>0.019105</td>\n",
       "      <td>0.139317</td>\n",
       "      <td>0.101675</td>\n",
       "      <td>...</td>\n",
       "      <td>0.141417</td>\n",
       "      <td>0.206333</td>\n",
       "      <td>-0.234480</td>\n",
       "      <td>-0.298500</td>\n",
       "      <td>0.132000</td>\n",
       "      <td>0.19100</td>\n",
       "      <td>-0.10000</td>\n",
       "      <td>0.017700</td>\n",
       "      <td>0.098075</td>\n",
       "      <td>0.053736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540610</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.049243</td>\n",
       "      <td>-0.008633</td>\n",
       "      <td>-0.05240</td>\n",
       "      <td>-0.035025</td>\n",
       "      <td>0.031200</td>\n",
       "      <td>0.0741</td>\n",
       "      <td>-0.005801</td>\n",
       "      <td>0.074317</td>\n",
       "      <td>-0.023225</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007568</td>\n",
       "      <td>0.196667</td>\n",
       "      <td>-0.077160</td>\n",
       "      <td>-0.095075</td>\n",
       "      <td>0.025267</td>\n",
       "      <td>0.00315</td>\n",
       "      <td>-0.26900</td>\n",
       "      <td>0.026992</td>\n",
       "      <td>0.060400</td>\n",
       "      <td>-0.030301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 16833 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer              15E1.2   2'-PDE       76P  \\\n",
       "GSM540550                    1.0 -0.091366 -0.018767 -0.09565 -0.177225   \n",
       "GSM540551                    1.0 -0.009954 -0.036233 -0.03820 -0.016975   \n",
       "GSM540552                    1.0  0.064026  0.126533 -0.27350 -0.015500   \n",
       "GSM540553                    1.0  0.223229  0.015967 -0.18850 -0.230250   \n",
       "GSM540554                    1.0  0.015514 -0.001400  0.07720 -0.381500   \n",
       "...                          ...       ...       ...      ...       ...   \n",
       "GSM540606                    0.0  0.135557  0.011360  0.01900 -0.010340   \n",
       "GSM540607                    0.0  0.125829 -0.109767 -0.11105 -0.110975   \n",
       "GSM540608                    0.0  0.009514  0.300333  0.18200  0.131200   \n",
       "GSM540609                    0.0  0.174671  0.013167 -0.03220 -0.107950   \n",
       "GSM540610                    0.0  0.049243 -0.008633 -0.05240 -0.035025   \n",
       "\n",
       "                7A5    A1BG     A2BP1       A2M     A2ML1  ...   ZYG11BL  \\\n",
       "GSM540550 -0.060111  0.2610 -0.110698 -0.071700 -0.056388  ...  0.118914   \n",
       "GSM540551  0.072104  0.0237  0.088163 -0.075717 -0.025460  ...  0.128233   \n",
       "GSM540552  0.055131  0.1170 -0.057415  0.184000  0.121440  ...  0.077617   \n",
       "GSM540553 -0.020300  0.1540  0.021842  0.176317  0.166813  ... -0.034995   \n",
       "GSM540554  0.201556  0.1410  0.034802  0.034750  0.002068  ... -0.074967   \n",
       "...             ...     ...       ...       ...       ...  ...       ...   \n",
       "GSM540606  0.079336 -0.0340  0.066139  0.115183 -0.013030  ... -0.131533   \n",
       "GSM540607 -0.004712  0.0540  0.069542  0.112083 -0.029762  ... -0.153200   \n",
       "GSM540608 -0.077400 -0.1540  0.006996  0.168233  0.093887  ...  0.372533   \n",
       "GSM540609 -0.000033  0.2420  0.019105  0.139317  0.101675  ...  0.141417   \n",
       "GSM540610  0.031200  0.0741 -0.005801  0.074317 -0.023225  ... -0.007568   \n",
       "\n",
       "                ZYX     ZZEF1      ZZZ3  bA16L21.2.1  bA9F11.1  dJ341D10.1  \\\n",
       "GSM540550 -0.034700 -0.129433 -0.079166     0.029367   0.15015     0.00661   \n",
       "GSM540551  0.021967 -0.158727 -0.043763     0.102400   0.09530    -0.16100   \n",
       "GSM540552  0.089867 -0.175639 -0.262375    -0.004233  -0.01800    -0.12000   \n",
       "GSM540553  0.084667 -0.123181 -0.159617    -0.040867   0.26300    -0.13300   \n",
       "GSM540554 -0.098300 -0.021133  0.018503    -0.086867   0.03475    -0.13800   \n",
       "...             ...       ...       ...          ...       ...         ...   \n",
       "GSM540606  0.057200 -0.026909 -0.087575     0.167833  -0.15180    -0.12400   \n",
       "GSM540607  0.041467 -0.100020 -0.126593     0.107967  -0.28300    -0.27100   \n",
       "GSM540608  0.072733  0.083072 -0.300383    -0.159400  -0.07216    -0.17600   \n",
       "GSM540609  0.206333 -0.234480 -0.298500     0.132000   0.19100    -0.10000   \n",
       "GSM540610  0.196667 -0.077160 -0.095075     0.025267   0.00315    -0.26900   \n",
       "\n",
       "            hCAP-D3   hCAP-H2   mimitin  \n",
       "GSM540550 -0.104746  0.053800 -0.052605  \n",
       "GSM540551 -0.010522  0.027662  0.082836  \n",
       "GSM540552  0.026106  0.011612  0.032056  \n",
       "GSM540553  0.072607  0.063725 -0.053923  \n",
       "GSM540554 -0.014310 -0.114200 -0.005092  \n",
       "...             ...       ...       ...  \n",
       "GSM540606  0.054733 -0.071883 -0.019704  \n",
       "GSM540607  0.012011 -0.031462  0.012892  \n",
       "GSM540608  0.007011  0.068550 -0.295004  \n",
       "GSM540609  0.017700  0.098075  0.053736  \n",
       "GSM540610  0.026992  0.060400 -0.030301  \n",
       "\n",
       "[61 rows x 16833 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data.drop(merged_data.columns[1], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adrenocortical Cancer</th>\n",
       "      <th>15E1.2</th>\n",
       "      <th>2'-PDE</th>\n",
       "      <th>76P</th>\n",
       "      <th>7A5</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2BP1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZYG11BL</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>bA16L21.2.1</th>\n",
       "      <th>bA9F11.1</th>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <th>hCAP-D3</th>\n",
       "      <th>hCAP-H2</th>\n",
       "      <th>mimitin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM540550</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.018767</td>\n",
       "      <td>-0.09565</td>\n",
       "      <td>-0.177225</td>\n",
       "      <td>-0.060111</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.110698</td>\n",
       "      <td>-0.071700</td>\n",
       "      <td>-0.056388</td>\n",
       "      <td>0.115900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.118914</td>\n",
       "      <td>-0.034700</td>\n",
       "      <td>-0.129433</td>\n",
       "      <td>-0.079166</td>\n",
       "      <td>0.029367</td>\n",
       "      <td>0.15015</td>\n",
       "      <td>0.00661</td>\n",
       "      <td>-0.104746</td>\n",
       "      <td>0.053800</td>\n",
       "      <td>-0.052605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540551</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.036233</td>\n",
       "      <td>-0.03820</td>\n",
       "      <td>-0.016975</td>\n",
       "      <td>0.072104</td>\n",
       "      <td>0.0237</td>\n",
       "      <td>0.088163</td>\n",
       "      <td>-0.075717</td>\n",
       "      <td>-0.025460</td>\n",
       "      <td>0.056557</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128233</td>\n",
       "      <td>0.021967</td>\n",
       "      <td>-0.158727</td>\n",
       "      <td>-0.043763</td>\n",
       "      <td>0.102400</td>\n",
       "      <td>0.09530</td>\n",
       "      <td>-0.16100</td>\n",
       "      <td>-0.010522</td>\n",
       "      <td>0.027662</td>\n",
       "      <td>0.082836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540552</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.126533</td>\n",
       "      <td>-0.27350</td>\n",
       "      <td>-0.015500</td>\n",
       "      <td>0.055131</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>-0.057415</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>0.121440</td>\n",
       "      <td>-0.057271</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077617</td>\n",
       "      <td>0.089867</td>\n",
       "      <td>-0.175639</td>\n",
       "      <td>-0.262375</td>\n",
       "      <td>-0.004233</td>\n",
       "      <td>-0.01800</td>\n",
       "      <td>-0.12000</td>\n",
       "      <td>0.026106</td>\n",
       "      <td>0.011612</td>\n",
       "      <td>0.032056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540553</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.015967</td>\n",
       "      <td>-0.18850</td>\n",
       "      <td>-0.230250</td>\n",
       "      <td>-0.020300</td>\n",
       "      <td>0.1540</td>\n",
       "      <td>0.021842</td>\n",
       "      <td>0.176317</td>\n",
       "      <td>0.166813</td>\n",
       "      <td>0.139629</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.034995</td>\n",
       "      <td>0.084667</td>\n",
       "      <td>-0.123181</td>\n",
       "      <td>-0.159617</td>\n",
       "      <td>-0.040867</td>\n",
       "      <td>0.26300</td>\n",
       "      <td>-0.13300</td>\n",
       "      <td>0.072607</td>\n",
       "      <td>0.063725</td>\n",
       "      <td>-0.053923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540554</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.001400</td>\n",
       "      <td>0.07720</td>\n",
       "      <td>-0.381500</td>\n",
       "      <td>0.201556</td>\n",
       "      <td>0.1410</td>\n",
       "      <td>0.034802</td>\n",
       "      <td>0.034750</td>\n",
       "      <td>0.002068</td>\n",
       "      <td>-0.060130</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.074967</td>\n",
       "      <td>-0.098300</td>\n",
       "      <td>-0.021133</td>\n",
       "      <td>0.018503</td>\n",
       "      <td>-0.086867</td>\n",
       "      <td>0.03475</td>\n",
       "      <td>-0.13800</td>\n",
       "      <td>-0.014310</td>\n",
       "      <td>-0.114200</td>\n",
       "      <td>-0.005092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540606</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011360</td>\n",
       "      <td>0.01900</td>\n",
       "      <td>-0.010340</td>\n",
       "      <td>0.079336</td>\n",
       "      <td>-0.0340</td>\n",
       "      <td>0.066139</td>\n",
       "      <td>0.115183</td>\n",
       "      <td>-0.013030</td>\n",
       "      <td>-0.120486</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.131533</td>\n",
       "      <td>0.057200</td>\n",
       "      <td>-0.026909</td>\n",
       "      <td>-0.087575</td>\n",
       "      <td>0.167833</td>\n",
       "      <td>-0.15180</td>\n",
       "      <td>-0.12400</td>\n",
       "      <td>0.054733</td>\n",
       "      <td>-0.071883</td>\n",
       "      <td>-0.019704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540607</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.109767</td>\n",
       "      <td>-0.11105</td>\n",
       "      <td>-0.110975</td>\n",
       "      <td>-0.004712</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>0.069542</td>\n",
       "      <td>0.112083</td>\n",
       "      <td>-0.029762</td>\n",
       "      <td>-0.127914</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.153200</td>\n",
       "      <td>0.041467</td>\n",
       "      <td>-0.100020</td>\n",
       "      <td>-0.126593</td>\n",
       "      <td>0.107967</td>\n",
       "      <td>-0.28300</td>\n",
       "      <td>-0.27100</td>\n",
       "      <td>0.012011</td>\n",
       "      <td>-0.031462</td>\n",
       "      <td>0.012892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540608</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.300333</td>\n",
       "      <td>0.18200</td>\n",
       "      <td>0.131200</td>\n",
       "      <td>-0.077400</td>\n",
       "      <td>-0.1540</td>\n",
       "      <td>0.006996</td>\n",
       "      <td>0.168233</td>\n",
       "      <td>0.093887</td>\n",
       "      <td>0.012549</td>\n",
       "      <td>...</td>\n",
       "      <td>0.372533</td>\n",
       "      <td>0.072733</td>\n",
       "      <td>0.083072</td>\n",
       "      <td>-0.300383</td>\n",
       "      <td>-0.159400</td>\n",
       "      <td>-0.07216</td>\n",
       "      <td>-0.17600</td>\n",
       "      <td>0.007011</td>\n",
       "      <td>0.068550</td>\n",
       "      <td>-0.295004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540609</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013167</td>\n",
       "      <td>-0.03220</td>\n",
       "      <td>-0.107950</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>0.2420</td>\n",
       "      <td>0.019105</td>\n",
       "      <td>0.139317</td>\n",
       "      <td>0.101675</td>\n",
       "      <td>0.150286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.141417</td>\n",
       "      <td>0.206333</td>\n",
       "      <td>-0.234480</td>\n",
       "      <td>-0.298500</td>\n",
       "      <td>0.132000</td>\n",
       "      <td>0.19100</td>\n",
       "      <td>-0.10000</td>\n",
       "      <td>0.017700</td>\n",
       "      <td>0.098075</td>\n",
       "      <td>0.053736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM540610</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.008633</td>\n",
       "      <td>-0.05240</td>\n",
       "      <td>-0.035025</td>\n",
       "      <td>0.031200</td>\n",
       "      <td>0.0741</td>\n",
       "      <td>-0.005801</td>\n",
       "      <td>0.074317</td>\n",
       "      <td>-0.023225</td>\n",
       "      <td>-0.038199</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.007568</td>\n",
       "      <td>0.196667</td>\n",
       "      <td>-0.077160</td>\n",
       "      <td>-0.095075</td>\n",
       "      <td>0.025267</td>\n",
       "      <td>0.00315</td>\n",
       "      <td>-0.26900</td>\n",
       "      <td>0.026992</td>\n",
       "      <td>0.060400</td>\n",
       "      <td>-0.030301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 16832 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Adrenocortical Cancer    15E1.2   2'-PDE       76P       7A5  \\\n",
       "GSM540550                    1.0 -0.018767 -0.09565 -0.177225 -0.060111   \n",
       "GSM540551                    1.0 -0.036233 -0.03820 -0.016975  0.072104   \n",
       "GSM540552                    1.0  0.126533 -0.27350 -0.015500  0.055131   \n",
       "GSM540553                    1.0  0.015967 -0.18850 -0.230250 -0.020300   \n",
       "GSM540554                    1.0 -0.001400  0.07720 -0.381500  0.201556   \n",
       "...                          ...       ...      ...       ...       ...   \n",
       "GSM540606                    0.0  0.011360  0.01900 -0.010340  0.079336   \n",
       "GSM540607                    0.0 -0.109767 -0.11105 -0.110975 -0.004712   \n",
       "GSM540608                    0.0  0.300333  0.18200  0.131200 -0.077400   \n",
       "GSM540609                    0.0  0.013167 -0.03220 -0.107950 -0.000033   \n",
       "GSM540610                    0.0 -0.008633 -0.05240 -0.035025  0.031200   \n",
       "\n",
       "             A1BG     A2BP1       A2M     A2ML1    A4GALT  ...   ZYG11BL  \\\n",
       "GSM540550  0.2610 -0.110698 -0.071700 -0.056388  0.115900  ...  0.118914   \n",
       "GSM540551  0.0237  0.088163 -0.075717 -0.025460  0.056557  ...  0.128233   \n",
       "GSM540552  0.1170 -0.057415  0.184000  0.121440 -0.057271  ...  0.077617   \n",
       "GSM540553  0.1540  0.021842  0.176317  0.166813  0.139629  ... -0.034995   \n",
       "GSM540554  0.1410  0.034802  0.034750  0.002068 -0.060130  ... -0.074967   \n",
       "...           ...       ...       ...       ...       ...  ...       ...   \n",
       "GSM540606 -0.0340  0.066139  0.115183 -0.013030 -0.120486  ... -0.131533   \n",
       "GSM540607  0.0540  0.069542  0.112083 -0.029762 -0.127914  ... -0.153200   \n",
       "GSM540608 -0.1540  0.006996  0.168233  0.093887  0.012549  ...  0.372533   \n",
       "GSM540609  0.2420  0.019105  0.139317  0.101675  0.150286  ...  0.141417   \n",
       "GSM540610  0.0741 -0.005801  0.074317 -0.023225 -0.038199  ... -0.007568   \n",
       "\n",
       "                ZYX     ZZEF1      ZZZ3  bA16L21.2.1  bA9F11.1  dJ341D10.1  \\\n",
       "GSM540550 -0.034700 -0.129433 -0.079166     0.029367   0.15015     0.00661   \n",
       "GSM540551  0.021967 -0.158727 -0.043763     0.102400   0.09530    -0.16100   \n",
       "GSM540552  0.089867 -0.175639 -0.262375    -0.004233  -0.01800    -0.12000   \n",
       "GSM540553  0.084667 -0.123181 -0.159617    -0.040867   0.26300    -0.13300   \n",
       "GSM540554 -0.098300 -0.021133  0.018503    -0.086867   0.03475    -0.13800   \n",
       "...             ...       ...       ...          ...       ...         ...   \n",
       "GSM540606  0.057200 -0.026909 -0.087575     0.167833  -0.15180    -0.12400   \n",
       "GSM540607  0.041467 -0.100020 -0.126593     0.107967  -0.28300    -0.27100   \n",
       "GSM540608  0.072733  0.083072 -0.300383    -0.159400  -0.07216    -0.17600   \n",
       "GSM540609  0.206333 -0.234480 -0.298500     0.132000   0.19100    -0.10000   \n",
       "GSM540610  0.196667 -0.077160 -0.095075     0.025267   0.00315    -0.26900   \n",
       "\n",
       "            hCAP-D3   hCAP-H2   mimitin  \n",
       "GSM540550 -0.104746  0.053800 -0.052605  \n",
       "GSM540551 -0.010522  0.027662  0.082836  \n",
       "GSM540552  0.026106  0.011612  0.032056  \n",
       "GSM540553  0.072607  0.063725 -0.053923  \n",
       "GSM540554 -0.014310 -0.114200 -0.005092  \n",
       "...             ...       ...       ...  \n",
       "GSM540606  0.054733 -0.071883 -0.019704  \n",
       "GSM540607  0.012011 -0.031462  0.012892  \n",
       "GSM540608  0.007011  0.068550 -0.295004  \n",
       "GSM540609  0.017700  0.098075  0.053736  \n",
       "GSM540610  0.026992  0.060400 -0.030301  \n",
       "\n",
       "[61 rows x 16832 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 61 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Adrenocortical Cancer', the least common label is '1.0' with 29 occurrences. This represents 47.54% of the dataset.\n",
      "The distribution of the feature 'Adrenocortical Cancer' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE21660</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>61</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "0  GSE21660       True          True      False    False       False   \n",
       "\n",
       "   sample_size note  \n",
       "0           61       "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.295070023Z",
     "start_time": "2024-01-10T21:40:21.338515691Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +: 'NoneType' and 'str'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[73], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m merged_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, \u001b[43mcondition_best_cohort\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m))\n\u001b[0;32m      2\u001b[0m merged_data\u001b[38;5;241m.\u001b[39mhead()\n",
      "\u001b[1;31mTypeError\u001b[0m: unsupported operand type(s) for +: 'NoneType' and 'str'"
     ]
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.358779256Z",
     "start_time": "2024-01-10T21:40:27.271337819Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n",
    "\n",
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546590575Z",
     "start_time": "2024-01-10T21:40:27.298495045Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546733889Z",
     "start_time": "2024-01-10T21:40:27.535249532Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Select appropriate models based on whether the dataset has batch effect.\n",
    "# We experiment on two models for each branch. We will decide which one to choose later.\n",
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:30.570471636Z",
     "start_time": "2024-01-10T21:40:27.546295159Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:47.575878022Z",
     "start_time": "2024-01-10T21:40:30.570907992Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.663737008Z",
     "start_time": "2024-01-10T21:40:47.576021186Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.673565176Z",
     "start_time": "2024-01-10T21:40:51.665291317Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.730825913Z",
     "start_time": "2024-01-10T21:40:51.675049728Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
