{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "# Set your preferred name\n",
    "USER = \"Jiayi\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'\n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output'\n",
    "TRAIT = 'Ankylosing Spondylitis'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.928143012Z",
     "start_time": "2024-01-10T21:37:25.918685567Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.554597756Z",
     "start_time": "2024-01-10T21:37:26.530982646Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.999521704Z",
     "start_time": "2024-01-10T21:37:26.987471072Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Adrenocortical_Cancer_(ACC)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_ACC_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.180581225Z",
     "start_time": "2024-01-10T21:37:27.217363685Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.185775284Z",
     "start_time": "2024-01-10T21:37:30.181745846Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'additional_pharmaceutical_therapy',\n",
       " 'additional_radiation_therapy',\n",
       " 'age_at_initial_pathologic_diagnosis',\n",
       " 'atypical_mitotic_figures',\n",
       " 'bcr_followup_barcode']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.269854441Z",
     "start_time": "2024-01-10T21:37:30.184225525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'age_at_initial_pathologic_diagnosis', 'atypical_mitotic_figures', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'clinical_M', 'ct_scan_findings', 'cytoplasm_presence_less_than_equal_25_percent', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'diffuse_architecture', 'distant_metastasis_anatomic_site', 'excess_adrenal_hormone_diagnosis_method_type', 'excess_adrenal_hormone_history_type', 'form_completion_date', 'gender', 'germline_testing_performed', 'histologic_disease_progression_present_indicator', 'histological_type', 'history_of_neoadjuvant_treatment', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'initial_weight', 'invasion_of_tumor_capsule', 'is_ffpe', 'laterality', 'lost_follow_up', 'lymph_node_examined_count', 'metastatic_neoplasm_confirmed_diagnosis_method_name', 'metastatic_neoplasm_confirmed_diagnosis_method_text', 'mitoses_count', 'mitotane_therapy', 'mitotane_therapy_adjuvant_setting', 'mitotane_therapy_for_macroscopic_residual_disease', 'mitotic_rate', 'necrosis', 'new_neoplasm_confirmed_diagnosis_method_name', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'nuclear_grade_III_IV', 'number_of_lymphnodes_positive_by_he', 'oct_embedded', 'other_dx', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'person_neoplasm_cancer_status', 'post_surgical_procedure_assessment_thyroid_gland_carcinoma_stats', 'postoperative_rx_tx', 'primary_lymph_node_presentation_assessment', 'primary_therapy_outcome_success', 'radiation_therapy', 'residual_tumor', 'ret', 'sample_type', 'sample_type_id', 'sinusoid_invasion', 'therapeutic_mitotane_levels_achieved', 'therapeutic_mitotane_lvl_macroscopic_residual', 'therapeutic_mitotane_lvl_progression', 'therapeutic_mitotane_lvl_recurrence', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'vial_number', 'vital_status', 'weiss_score', 'weiss_venous_invasion', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_ACC_mutation_curated_bcm_gene', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_percentile', '_GENOMIC_ID_data/public/TCGA/ACC/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_ACC_miRNA_HiSeq', '_GENOMIC_ID_TCGA_ACC_RPPA', '_GENOMIC_ID_TCGA_ACC_hMethyl450', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2', '_GENOMIC_ID_TCGA_ACC_gistic2thd', '_GENOMIC_ID_TCGA_ACC_PDMRNAseq', '_GENOMIC_ID_TCGA_ACC_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_ACC_gistic2', '_GENOMIC_ID_TCGA_ACC_mutation_bcm_gene', '_GENOMIC_ID_TCGA_ACC_mutation_curated_broad_gene']\\n\""
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270036796Z",
     "start_time": "2024-01-10T21:37:30.227180075Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = [ 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = [ 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270232386Z",
     "start_time": "2024-01-10T21:37:30.227282564Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'age_at_initial_pathologic_diagnosis': [58, 44, 23, 23, 30],\n",
       " 'days_to_birth': [-21496, -16090, -8624, -8451, -11171],\n",
       " 'year_of_initial_pathologic_diagnosis': [2000, 2004, 2008, 2000, 2000]}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270339062Z",
     "start_time": "2024-01-10T21:37:30.227412883Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270520195Z",
     "start_time": "2024-01-10T21:37:30.227467704Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gender': ['MALE', 'FEMALE', 'FEMALE', 'FEMALE', 'MALE']}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270626721Z",
     "start_time": "2024-01-10T21:37:30.227530279Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270719762Z",
     "start_time": "2024-01-10T21:37:30.227585330Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.001977877Z",
     "start_time": "2024-01-10T21:37:30.227635342Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12 input query terms found dup hits:\t[('GTF2IP1', 2), ('RBMY1A3P', 3), ('RPL31P11', 2), ('HERC2P2', 3), ('WASH3P', 3), ('NUDT9P1', 2), ('\n",
      "154 input query terms found no hit:\t['C16orf13', 'C16orf11', 'LOC100272146', 'LOC339240', 'NACAP1', 'LOC441204', 'KLRA1', 'FAM183A', 'FA\n",
      "10 input query terms found dup hits:\t[('SUGT1P1', 2), ('PTPRVP', 2), ('SNORA62', 3), ('IFITM4P', 7), ('HLA-DRB6', 2), ('FUNDC2P2', 2), ('\n",
      "190 input query terms found no hit:\t['NARFL', 'NFKBIL2', 'LOC150197', 'TMEM84', 'LOC162632', 'PPPDE1', 'PPPDE2', 'C1orf38', 'C1orf31', '\n",
      "11 input query terms found dup hits:\t[('PIP5K1P1', 2), ('HBD', 2), ('PPP1R2P1', 9), ('HSD17B7P2', 2), ('RPSAP9', 2), ('SNORD68', 2), ('SN\n",
      "149 input query terms found no hit:\t['FAM153C', 'C9orf167', 'CLK2P', 'CCDC76', 'CCDC75', 'CCDC72', 'HIST3H2BB', 'PRAC', 'LOC285780', 'LO\n",
      "15 input query terms found dup hits:\t[('SNORD58C', 2), ('UOX', 2), ('UBE2Q2P1', 3), ('PPP4R1L', 2), ('SNORD63', 3), ('ESPNP', 2), ('HBBP1\n",
      "158 input query terms found no hit:\t['C16orf91', 'LOC100130581', 'SFRS2', 'C16orf5', 'C16orf7', 'C16orf3', 'C21orf7', 'C21orf2', 'LOC389\n",
      "15 input query terms found dup hits:\t[('FAM66D', 3), ('FAM66A', 2), ('THSD1P1', 2), ('EEF1DP3', 2), ('PGM5P2', 2), ('UBE2MP1', 2), ('HAR1\n",
      "169 input query terms found no hit:\t['LOC284551', 'LOC285548', 'LOC728410', 'LOC541473', 'DULLARD', 'KIAA0368', 'EFTUD1', 'TWISTNB', 'SF\n",
      "13 input query terms found dup hits:\t[('S100A7L2', 2), ('POM121L8P', 2), ('MEG8', 2), ('KIR3DX1', 5), ('RFPL1S', 2), ('SNORD91B', 2), ('C\n",
      "165 input query terms found no hit:\t['TMEM188', 'PDZD3', 'FAM102B', 'FAM102A', 'SMCR7L', 'G6PC', 'OSTCL', 'LOC653544', 'LOC653545', 'USP\n",
      "16 input query terms found dup hits:\t[('PCNAP1', 2), ('SNORA63', 6), ('SERHL', 2), ('CEACAM22P', 2), ('SNORA16A', 2), ('FAM41AY1', 2), ('\n",
      "147 input query terms found no hit:\t['LRRC37A4', 'LOC100131726', 'CPSF3L', 'COL4A3BP', 'PAR1', 'LOC92973', 'MICALCL', 'SMCR7', 'HIST4H4'\n",
      "15 input query terms found dup hits:\t[('MBL1P', 2), ('SDHAP3', 2), ('PSORS1C3', 8), ('MYADML', 2), ('POM121L10P', 2), ('HLA-J', 9), ('HLA\n",
      "153 input query terms found no hit:\t['LOC100131193', 'KIAA1109', 'DUS2L', 'MYST1', 'IGLL3', 'C12orf69', 'NEURL', 'FAM188B', 'FLJ32063', \n",
      "9 input query terms found dup hits:\t[('PLGLA', 2), ('PA2G4P4', 2), ('RPL23AP7', 2), ('RPL23AP64', 3), ('RPSAP52', 2), ('TPRXL', 2), ('DP\n",
      "177 input query terms found no hit:\t['LOC283314', 'RBM9', 'C20orf118', 'LOC150786', 'C20orf112', 'C20orf111', 'C20orf117', 'C20orf114', \n",
      "18 input query terms found dup hits:\t[('TPTE2P1', 2), ('FOLH1B', 2), ('ID2B', 2), ('MSX2P1', 2), ('RAET1K', 2), ('BMS1P4', 2), ('RPL23AP8\n",
      "175 input query terms found no hit:\t['MYEOV2', 'FAM35B2', 'C20orf85', 'FAM175A', 'C17orf73', 'C17orf72', 'C17orf71', 'C17orf70', 'C17orf\n",
      "19 input query terms found dup hits:\t[('CATSPER2P1', 2), ('AQP7P1', 2), ('ATP8B5P', 2), ('OR7E91P', 2), ('ZNF204P', 2), ('CECR7', 2), ('M\n",
      "144 input query terms found no hit:\t['BASE', 'C14orf21', 'C14orf23', 'FAM5B', 'TROVE2', 'C18orf62', 'LOC401052', 'GNB2L1', 'FAM19A4', 'F\n",
      "13 input query terms found dup hits:\t[('P2RX6P', 2), ('RP9P', 2), ('NRADDP', 2), ('MORF4', 2), ('SNORA75', 6), ('SNORA72', 2), ('FMO6P', \n",
      "168 input query terms found no hit:\t['TTC15', 'C4orf29', 'FAM48A', 'C4orf21', 'C3orf65', 'C4orf23', 'C3orf67', 'LOC144742', 'C3orf63', '\n",
      "19 input query terms found dup hits:\t[('RPL13AP17', 2), ('DNAJB3', 2), ('SNORA58', 2), ('PMCHL1', 2), ('FAM95B1', 2), ('TMEM191A', 2), ('\n",
      "144 input query terms found no hit:\t['LOC254312', 'ZNF542', 'CSDAP1', 'FAM73A', 'FAM73B', 'TCTE3', 'ZNF37B', 'LOC728875', 'KIAA1826', 'P\n",
      "15 input query terms found dup hits:\t[('ZFP91-CNTF', 2), ('PRSS30P', 3), ('GSTM2P1', 2), ('C3P1', 2), ('NCF1B', 2), ('SNORD114-4', 2), ('\n",
      "138 input query terms found no hit:\t['LOC100129716', 'C14orf184', 'C14orf181', 'C14orf183', 'C14orf182', 'C7orf26', 'C7orf27', 'C7orf23'\n",
      "14 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P3', 2), ('ANXA2P2', 2), ('ZNF286B', 2), ('SAA3P', 2), ('GEMIN8P4', 2), ('DI\n",
      "132 input query terms found no hit:\t['AKR1CL1', 'CCDC21', 'CCDC23', 'TMEM111', 'APOB48R', 'LOC595101', 'KIAA1430', 'KIAA1432', 'ORC4L', \n",
      "11 input query terms found dup hits:\t[('TCP10L2', 2), ('FTLP10', 2), ('TPTE2P3', 2), ('VENTXP1', 2), ('XKRY', 2), ('HTR7P1', 2), ('STAG3L\n",
      "175 input query terms found no hit:\t['DFNB59', 'LOC100170939', 'FAM36A', 'C2orf70', 'C2orf71', 'C2orf77', 'C2orf79', 'LOC100128191', 'HR\n",
      "16 input query terms found dup hits:\t[('ZNF702P', 2), ('BRD7P3', 2), ('SCARNA17', 2), ('GABARAPL3', 2), ('MT1IP', 2), ('ADAM6', 3), ('PDZ\n",
      "142 input query terms found no hit:\t['C9orf30', 'PEG3AS', 'C9orf37', 'TXNDC3', 'LOC442308', 'FAM46D', 'FAM46A', 'FAM46B', 'FAM46C', 'WDR\n",
      "10 input query terms found dup hits:\t[('FAM197Y2', 3), ('SNORA71B', 2), ('PRKY', 2), ('NAT8B', 3), ('VENTXP7', 2), ('SPRR2C', 2), ('RPL23\n",
      "148 input query terms found no hit:\t['FCGR1B', 'RG9MTD2', 'RARS', 'LOC727677', 'CCRN4L', 'SEPT14', 'HKR1', 'C1orf112', 'LOC100270710', '\n",
      "16 input query terms found dup hits:\t[('OR6W1P', 3), ('LYPLA2P1', 6), ('RPL23AP53', 2), ('TRPC2', 2), ('RBMY2FP', 2), ('BAGE2', 2), ('POM\n",
      "150 input query terms found no hit:\t['SIP1', 'FAM129C', 'GIF', 'TMCO7', 'CXorf61', 'CXorf64', 'ZNF192', 'SEPP1', 'KIRREL', 'MGC16121', '\n",
      "17 input query terms found dup hits:\t[('ISCA1P1', 2), ('PCDHB19P', 3), ('NAPSB', 2), ('WASH7P', 2), ('TTTY11', 2), ('HCG4', 9), ('OR5E1P'\n",
      "135 input query terms found no hit:\t['PPYR1', 'FAM82A2', 'FLJ37453', 'FAM82A1', 'KIAA0947', 'MOBKL2B', 'MOBKL2C', 'MOBKL2A', 'CMAH', 'GR\n",
      "5 input query terms found dup hits:\t[('GGNBP1', 2), ('WASH5P', 2), ('RPL29P2', 2), ('CDR1', 2), ('OR10J3', 3)]\n",
      "78 input query terms found no hit:\t['C14orf174', 'C14orf176', 'C14orf177', 'C14orf179', 'PTPLB', 'SGEF', 'LOC400752', 'GRLF1', 'SPANXE'\n"
     ]
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.108889749Z",
     "start_time": "2024-01-10T21:38:18.021199718Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.641092</td>\n",
       "      <td>-0.325826</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.266428</td>\n",
       "      <td>0.355422</td>\n",
       "      <td>0.03719</td>\n",
       "      <td>0.243706</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.520186</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.182978</td>\n",
       "      <td>-0.615817</td>\n",
       "      <td>-0.281533</td>\n",
       "      <td>3.02111</td>\n",
       "      <td>-0.927577</td>\n",
       "      <td>-1.006227</td>\n",
       "      <td>1.119905</td>\n",
       "      <td>-2.185533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.864792</td>\n",
       "      <td>2.766674</td>\n",
       "      <td>0.321165</td>\n",
       "      <td>1.000728</td>\n",
       "      <td>0.836122</td>\n",
       "      <td>0.35439</td>\n",
       "      <td>-0.436694</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.318586</td>\n",
       "      <td>1.056018</td>\n",
       "      <td>0.393822</td>\n",
       "      <td>2.366583</td>\n",
       "      <td>-0.955033</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>1.020723</td>\n",
       "      <td>1.226373</td>\n",
       "      <td>1.164005</td>\n",
       "      <td>0.265067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.723192</td>\n",
       "      <td>-0.362926</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.639828</td>\n",
       "      <td>-0.199578</td>\n",
       "      <td>-0.48331</td>\n",
       "      <td>0.143606</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.574486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>-0.113317</td>\n",
       "      <td>-3.803333</td>\n",
       "      <td>-0.61009</td>\n",
       "      <td>0.397623</td>\n",
       "      <td>-0.675227</td>\n",
       "      <td>1.196005</td>\n",
       "      <td>-3.161633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.576792</td>\n",
       "      <td>-2.086226</td>\n",
       "      <td>2.463765</td>\n",
       "      <td>1.382228</td>\n",
       "      <td>-1.115678</td>\n",
       "      <td>-1.23621</td>\n",
       "      <td>0.615806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.279486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.078622</td>\n",
       "      <td>1.095983</td>\n",
       "      <td>-0.908533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>0.661823</td>\n",
       "      <td>0.458273</td>\n",
       "      <td>0.839605</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J6-01</th>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.311992</td>\n",
       "      <td>5.225974</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.967928</td>\n",
       "      <td>-0.393778</td>\n",
       "      <td>-0.38231</td>\n",
       "      <td>-0.060194</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>1.607218</td>\n",
       "      <td>2.481122</td>\n",
       "      <td>-0.946617</td>\n",
       "      <td>-0.570533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.425177</td>\n",
       "      <td>0.938573</td>\n",
       "      <td>0.495005</td>\n",
       "      <td>-1.733333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Ankylosing Spondylitis  Age  Gender  ARHGEF10L     HIF3A  \\\n",
       "sampleID                                                                    \n",
       "TCGA-OR-A5J1-01                       1   58       1  -0.641092 -0.325826   \n",
       "TCGA-OR-A5J2-01                       1   44       0  -1.864792  2.766674   \n",
       "TCGA-OR-A5J3-01                       1   23       0  -0.723192 -0.362926   \n",
       "TCGA-OR-A5J5-01                       1   30       1  -1.576792 -2.086226   \n",
       "TCGA-OR-A5J6-01                       1   29       0  -2.311992  5.225974   \n",
       "\n",
       "                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n",
       "sampleID                                                          ...   \n",
       "TCGA-OR-A5J1-01 -0.531035  1.266428  0.355422  0.03719  0.243706  ...   \n",
       "TCGA-OR-A5J2-01  0.321165  1.000728  0.836122  0.35439 -0.436694  ...   \n",
       "TCGA-OR-A5J3-01 -0.531035  0.639828 -0.199578 -0.48331  0.143606  ...   \n",
       "TCGA-OR-A5J5-01  2.463765  1.382228 -1.115678 -1.23621  0.615806  ...   \n",
       "TCGA-OR-A5J6-01 -0.531035  0.967928 -0.393778 -0.38231 -0.060194  ...   \n",
       "\n",
       "                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n",
       "sampleID                                                                     \n",
       "TCGA-OR-A5J1-01 -1.520186 -0.086682 -0.182978 -0.615817 -0.281533  3.02111   \n",
       "TCGA-OR-A5J2-01 -0.318586  1.056018  0.393822  2.366583 -0.955033 -1.28139   \n",
       "TCGA-OR-A5J3-01 -0.574486 -0.086682 -0.748878 -0.113317 -3.803333 -0.61009   \n",
       "TCGA-OR-A5J5-01 -0.279486 -0.086682  0.078622  1.095983 -0.908533 -1.28139   \n",
       "TCGA-OR-A5J6-01 -2.090786  1.607218  2.481122 -0.946617 -0.570533 -1.28139   \n",
       "\n",
       "                    TULP3     BCL6B     GSTK1      SELP  \n",
       "sampleID                                                 \n",
       "TCGA-OR-A5J1-01 -0.927577 -1.006227  1.119905 -2.185533  \n",
       "TCGA-OR-A5J2-01  1.020723  1.226373  1.164005  0.265067  \n",
       "TCGA-OR-A5J3-01  0.397623 -0.675227  1.196005 -3.161633  \n",
       "TCGA-OR-A5J5-01  0.661823  0.458273  0.839605 -5.525533  \n",
       "TCGA-OR-A5J6-01 -0.425177  0.938573  0.495005 -1.733333  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.112793384Z",
     "start_time": "2024-01-10T21:38:18.107415526Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 79 samples.\n",
      "For the feature 'Ankylosing Spondylitis', the least common label is '1' with 79 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Ankylosing Spondylitis' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 35.0\n",
      "  50% (Median): 49.0\n",
      "  75%: 59.5\n",
      "Min: 14\n",
      "Max: 77\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1' with 31 occurrences. This represents 39.24% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.377898533Z",
     "start_time": "2024-01-10T21:38:18.111417452Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425336629Z",
     "start_time": "2024-01-10T21:38:31.419043732Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output\\Jiayi\\Ankylosing-Spondylitis\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE100648', 'GSE11886', 'GSE25101', 'GSE73754']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Ankylosing-Spondylitis\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:02.935079426Z",
     "start_time": "2024-01-10T21:39:02.923698385Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Early arthritis B cell gene expression profiling\"\n",
      "!Series_summary\t\"242 patients recruited from an early arthritis clinic donated RNA and DNA from freshly isolated and purified peripheral blood CD19+ B cells. Global gene expression measurement was carried out using Illumina BeadChip HT12v4 microarrays. Objectives included the identification of B cell transcripts differentially expressed between disease phenotypes, where all patients were naive to immunomodulatory therapy. In addition an eQTL analysis was carried out with reference to known genotype data for this cohort of patients\"\n",
      "!Series_overall_design\t\"Cross sectional; diagnoses confirmed at >2 years' follow-up.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2690230</th>\n",
       "      <th>GSM2690231</th>\n",
       "      <th>GSM2690232</th>\n",
       "      <th>GSM2690233</th>\n",
       "      <th>GSM2690234</th>\n",
       "      <th>GSM2690235</th>\n",
       "      <th>GSM2690236</th>\n",
       "      <th>GSM2690237</th>\n",
       "      <th>GSM2690238</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2690462</th>\n",
       "      <th>GSM2690463</th>\n",
       "      <th>GSM2690464</th>\n",
       "      <th>GSM2690465</th>\n",
       "      <th>GSM2690466</th>\n",
       "      <th>GSM2690467</th>\n",
       "      <th>GSM2690468</th>\n",
       "      <th>GSM2690469</th>\n",
       "      <th>GSM2690470</th>\n",
       "      <th>GSM2690471</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "      <td>cell type: CD19+ B cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>first_diagnosis: Non-Inflammatory</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Osteoarthritis</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Enteropathic Arthritis</td>\n",
       "      <td>first_diagnosis: Other Inflammatory Arthritis</td>\n",
       "      <td>...</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Undifferentiated Inflammatory...</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>first_diagnosis: Rheumatoid Arthritis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>working_diagnosis: Osteoarthritis</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>working_diagnosis: Non-Inflammatory</td>\n",
       "      <td>working_diagnosis: Reactive Arthritis</td>\n",
       "      <td>working_diagnosis: Osteoarthritis</td>\n",
       "      <td>working_diagnosis: Reactive Arthritis</td>\n",
       "      <td>working_diagnosis: Enteropathic Arthritis</td>\n",
       "      <td>working_diagnosis: Other Inflammatory Arthritis</td>\n",
       "      <td>...</td>\n",
       "      <td>working_diagnosis: Undifferentiated Inflammato...</td>\n",
       "      <td>working_diagnosis: Undifferentiated Inflammato...</td>\n",
       "      <td>working_diagnosis: Undifferentiated Inflammato...</td>\n",
       "      <td>working_diagnosis: Undifferentiated Inflammato...</td>\n",
       "      <td>working_diagnosis: Undifferentiated Inflammato...</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "      <td>working_diagnosis: Rheumatoid Arthritis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 243 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                               GSM2690230  \\\n",
       "0  !Sample_characteristics_ch1                 tissue: peripheral blood   \n",
       "1  !Sample_characteristics_ch1                 cell type: CD19+ B cells   \n",
       "2  !Sample_characteristics_ch1    first_diagnosis: Rheumatoid Arthritis   \n",
       "3  !Sample_characteristics_ch1  working_diagnosis: Rheumatoid Arthritis   \n",
       "4  !Sample_characteristics_ch1                                   Sex: M   \n",
       "\n",
       "                                          GSM2690231  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3                  working_diagnosis: Osteoarthritis   \n",
       "4                                             Sex: F   \n",
       "\n",
       "                                GSM2690232  \\\n",
       "0                 tissue: peripheral blood   \n",
       "1                 cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Rheumatoid Arthritis   \n",
       "3  working_diagnosis: Rheumatoid Arthritis   \n",
       "4                                   Sex: F   \n",
       "\n",
       "                            GSM2690233  \\\n",
       "0             tissue: peripheral blood   \n",
       "1             cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Non-Inflammatory   \n",
       "3  working_diagnosis: Non-Inflammatory   \n",
       "4                               Sex: F   \n",
       "\n",
       "                                          GSM2690234  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3              working_diagnosis: Reactive Arthritis   \n",
       "4                                             Sex: M   \n",
       "\n",
       "                          GSM2690235  \\\n",
       "0           tissue: peripheral blood   \n",
       "1           cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Osteoarthritis   \n",
       "3  working_diagnosis: Osteoarthritis   \n",
       "4                             Sex: F   \n",
       "\n",
       "                                          GSM2690236  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3              working_diagnosis: Reactive Arthritis   \n",
       "4                                             Sex: M   \n",
       "\n",
       "                                  GSM2690237  \\\n",
       "0                   tissue: peripheral blood   \n",
       "1                   cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Enteropathic Arthritis   \n",
       "3  working_diagnosis: Enteropathic Arthritis   \n",
       "4                                     Sex: F   \n",
       "\n",
       "                                        GSM2690238  ...  \\\n",
       "0                         tissue: peripheral blood  ...   \n",
       "1                         cell type: CD19+ B cells  ...   \n",
       "2    first_diagnosis: Other Inflammatory Arthritis  ...   \n",
       "3  working_diagnosis: Other Inflammatory Arthritis  ...   \n",
       "4                                           Sex: F  ...   \n",
       "\n",
       "                                          GSM2690462  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3  working_diagnosis: Undifferentiated Inflammato...   \n",
       "4                                             Sex: F   \n",
       "\n",
       "                                          GSM2690463  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3  working_diagnosis: Undifferentiated Inflammato...   \n",
       "4                                             Sex: F   \n",
       "\n",
       "                                          GSM2690464  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3  working_diagnosis: Undifferentiated Inflammato...   \n",
       "4                                             Sex: F   \n",
       "\n",
       "                                          GSM2690465  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3  working_diagnosis: Undifferentiated Inflammato...   \n",
       "4                                             Sex: F   \n",
       "\n",
       "                                          GSM2690466  \\\n",
       "0                           tissue: peripheral blood   \n",
       "1                           cell type: CD19+ B cells   \n",
       "2  first_diagnosis: Undifferentiated Inflammatory...   \n",
       "3  working_diagnosis: Undifferentiated Inflammato...   \n",
       "4                                             Sex: F   \n",
       "\n",
       "                                GSM2690467  \\\n",
       "0                 tissue: peripheral blood   \n",
       "1                 cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Rheumatoid Arthritis   \n",
       "3  working_diagnosis: Rheumatoid Arthritis   \n",
       "4                                   Sex: M   \n",
       "\n",
       "                                GSM2690468  \\\n",
       "0                 tissue: peripheral blood   \n",
       "1                 cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Rheumatoid Arthritis   \n",
       "3  working_diagnosis: Rheumatoid Arthritis   \n",
       "4                                   Sex: F   \n",
       "\n",
       "                                GSM2690469  \\\n",
       "0                 tissue: peripheral blood   \n",
       "1                 cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Rheumatoid Arthritis   \n",
       "3  working_diagnosis: Rheumatoid Arthritis   \n",
       "4                                   Sex: F   \n",
       "\n",
       "                                GSM2690470  \\\n",
       "0                 tissue: peripheral blood   \n",
       "1                 cell type: CD19+ B cells   \n",
       "2    first_diagnosis: Rheumatoid Arthritis   \n",
       "3  working_diagnosis: Rheumatoid Arthritis   \n",
       "4                                   Sex: F   \n",
       "\n",
       "                                GSM2690471  \n",
       "0                 tissue: peripheral blood  \n",
       "1                 cell type: CD19+ B cells  \n",
       "2    first_diagnosis: Rheumatoid Arthritis  \n",
       "3  working_diagnosis: Rheumatoid Arthritis  \n",
       "4                                   Sex: F  \n",
       "\n",
       "[5 rows x 243 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Completed\n",
    "cohort = accession_num = \"GSE100648\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'working_diagnosis: Rheumatoid Arthritis',\n",
       "       'working_diagnosis: Osteoarthritis',\n",
       "       'working_diagnosis: Non-Inflammatory',\n",
       "       'working_diagnosis: Reactive Arthritis',\n",
       "       'working_diagnosis: Enteropathic Arthritis',\n",
       "       'working_diagnosis: Other Inflammatory Arthritis',\n",
       "       'working_diagnosis: Psoriatic Arthritis',\n",
       "       'working_diagnosis: Undifferentiated Spondylo-Arthropathy',\n",
       "       'working_diagnosis: Crystal Arthritis',\n",
       "       'working_diagnosis: Unknown',\n",
       "       'working_diagnosis: Undifferentiated Inflammatory Arthritis',\n",
       "       'working_diagnosis: Ankylosing Spondylitis',\n",
       "       'working_diagnosis: Lupus/Other CTD-Associated'], dtype=object)"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[3]\n",
    "tumor_stage_row.unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 3\n",
    "age_row = None\n",
    "gender_row = 4\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'working_diagnosis: Ankylosing Spondylitis':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°çš„ç›®çš„æ˜¯å°†å¹´é¾„çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºä¸€ä¸ªè¿žç»­çš„æ•°å€¼åž‹è¡¨ç¤ºã€‚å¦‚æžœå¹´é¾„æœªçŸ¥ï¼ˆä¾‹å¦‚ï¼Œæ ‡è®°ä¸º'n.a.'ï¼‰ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# å‡½æ•°å°è¯•ä»Žä¼ å…¥çš„å­—ç¬¦ä¸²ä¸­æå–å‡ºä¸€ä¸ªæ•´æ•°ä½œä¸ºå¹´é¾„å€¼ã€‚å¦‚æžœå­—ç¬¦ä¸²çš„æ ¼å¼ä¸ç¬¦åˆé¢„æœŸï¼Œå¯¼è‡´æå–å¤±è´¥ï¼ŒåŒæ ·è¿”å›žNoneã€‚\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†æ€§åˆ«çš„å­—ç¬¦ä¸²è¡¨ç¤ºè½¬æ¢ä¸ºäºŒè¿›åˆ¶å€¼ï¼Œå…¶ä¸­â€œfemaleâ€å¯¹åº”1ï¼Œâ€œmaleâ€å¯¹åº”0ã€‚å¦‚æžœæ€§åˆ«æœªçŸ¥æˆ–å­—ç¬¦ä¸²ä¸ç¬¦åˆé¢„æœŸæ ¼å¼ï¼Œåˆ™è¿”å›žNoneã€‚\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_24124\\2239879167.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_24124\\2239879167.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2690230</th>\n",
       "      <th>GSM2690231</th>\n",
       "      <th>GSM2690232</th>\n",
       "      <th>GSM2690233</th>\n",
       "      <th>GSM2690234</th>\n",
       "      <th>GSM2690235</th>\n",
       "      <th>GSM2690236</th>\n",
       "      <th>GSM2690237</th>\n",
       "      <th>GSM2690238</th>\n",
       "      <th>GSM2690239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2690462</th>\n",
       "      <th>GSM2690463</th>\n",
       "      <th>GSM2690464</th>\n",
       "      <th>GSM2690465</th>\n",
       "      <th>GSM2690466</th>\n",
       "      <th>GSM2690467</th>\n",
       "      <th>GSM2690468</th>\n",
       "      <th>GSM2690469</th>\n",
       "      <th>GSM2690470</th>\n",
       "      <th>GSM2690471</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 242 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GSM2690230  GSM2690231  GSM2690232  GSM2690233  \\\n",
       "Ankylosing Spondylitis           0           0           0           0   \n",
       "Gender                           0           1           1           1   \n",
       "\n",
       "                        GSM2690234  GSM2690235  GSM2690236  GSM2690237  \\\n",
       "Ankylosing Spondylitis           0           0           0           0   \n",
       "Gender                           0           1           0           1   \n",
       "\n",
       "                        GSM2690238  GSM2690239  ...  GSM2690462  GSM2690463  \\\n",
       "Ankylosing Spondylitis           0           0  ...           0           0   \n",
       "Gender                           1           1  ...           1           1   \n",
       "\n",
       "                        GSM2690464  GSM2690465  GSM2690466  GSM2690467  \\\n",
       "Ankylosing Spondylitis           0           0           0           0   \n",
       "Gender                           1           1           1           0   \n",
       "\n",
       "                        GSM2690468  GSM2690469  GSM2690470  GSM2690471  \n",
       "Ankylosing Spondylitis           0           0           0           0  \n",
       "Gender                           1           1           1           1  \n",
       "\n",
       "[2 rows x 242 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2690230</th>\n",
       "      <th>GSM2690231</th>\n",
       "      <th>GSM2690232</th>\n",
       "      <th>GSM2690233</th>\n",
       "      <th>GSM2690234</th>\n",
       "      <th>GSM2690235</th>\n",
       "      <th>GSM2690236</th>\n",
       "      <th>GSM2690237</th>\n",
       "      <th>GSM2690238</th>\n",
       "      <th>GSM2690239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2690462</th>\n",
       "      <th>GSM2690463</th>\n",
       "      <th>GSM2690464</th>\n",
       "      <th>GSM2690465</th>\n",
       "      <th>GSM2690466</th>\n",
       "      <th>GSM2690467</th>\n",
       "      <th>GSM2690468</th>\n",
       "      <th>GSM2690469</th>\n",
       "      <th>GSM2690470</th>\n",
       "      <th>GSM2690471</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>14.415838</td>\n",
       "      <td>14.420469</td>\n",
       "      <td>14.454755</td>\n",
       "      <td>14.548113</td>\n",
       "      <td>14.393246</td>\n",
       "      <td>14.476773</td>\n",
       "      <td>14.536234</td>\n",
       "      <td>14.655704</td>\n",
       "      <td>14.495031</td>\n",
       "      <td>14.532721</td>\n",
       "      <td>...</td>\n",
       "      <td>14.649794</td>\n",
       "      <td>14.700669</td>\n",
       "      <td>14.376907</td>\n",
       "      <td>14.617264</td>\n",
       "      <td>14.512447</td>\n",
       "      <td>14.295207</td>\n",
       "      <td>14.482401</td>\n",
       "      <td>14.478883</td>\n",
       "      <td>14.605284</td>\n",
       "      <td>14.556862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1343295</th>\n",
       "      <td>12.501028</td>\n",
       "      <td>11.415616</td>\n",
       "      <td>11.695755</td>\n",
       "      <td>11.923042</td>\n",
       "      <td>11.554511</td>\n",
       "      <td>11.474396</td>\n",
       "      <td>12.061302</td>\n",
       "      <td>11.915192</td>\n",
       "      <td>11.952768</td>\n",
       "      <td>11.313382</td>\n",
       "      <td>...</td>\n",
       "      <td>11.580915</td>\n",
       "      <td>10.424846</td>\n",
       "      <td>10.646008</td>\n",
       "      <td>11.864454</td>\n",
       "      <td>11.133300</td>\n",
       "      <td>11.915684</td>\n",
       "      <td>11.448013</td>\n",
       "      <td>11.947450</td>\n",
       "      <td>11.322478</td>\n",
       "      <td>11.192065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651199</th>\n",
       "      <td>6.627739</td>\n",
       "      <td>6.607227</td>\n",
       "      <td>6.585315</td>\n",
       "      <td>6.676918</td>\n",
       "      <td>6.565506</td>\n",
       "      <td>6.706464</td>\n",
       "      <td>6.494084</td>\n",
       "      <td>6.584417</td>\n",
       "      <td>6.575962</td>\n",
       "      <td>6.658423</td>\n",
       "      <td>...</td>\n",
       "      <td>6.702769</td>\n",
       "      <td>6.486990</td>\n",
       "      <td>6.786289</td>\n",
       "      <td>6.622436</td>\n",
       "      <td>6.590972</td>\n",
       "      <td>6.481921</td>\n",
       "      <td>6.641028</td>\n",
       "      <td>6.683143</td>\n",
       "      <td>6.478434</td>\n",
       "      <td>6.537222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651209</th>\n",
       "      <td>6.556000</td>\n",
       "      <td>7.053035</td>\n",
       "      <td>6.802224</td>\n",
       "      <td>6.954742</td>\n",
       "      <td>6.890331</td>\n",
       "      <td>7.070821</td>\n",
       "      <td>6.954012</td>\n",
       "      <td>6.929918</td>\n",
       "      <td>6.918314</td>\n",
       "      <td>6.734604</td>\n",
       "      <td>...</td>\n",
       "      <td>6.800404</td>\n",
       "      <td>7.103074</td>\n",
       "      <td>6.853516</td>\n",
       "      <td>6.867996</td>\n",
       "      <td>6.674861</td>\n",
       "      <td>6.775554</td>\n",
       "      <td>6.766406</td>\n",
       "      <td>6.716863</td>\n",
       "      <td>6.870444</td>\n",
       "      <td>6.863519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651210</th>\n",
       "      <td>6.597989</td>\n",
       "      <td>6.552052</td>\n",
       "      <td>6.658103</td>\n",
       "      <td>6.666267</td>\n",
       "      <td>6.727378</td>\n",
       "      <td>6.667242</td>\n",
       "      <td>6.653112</td>\n",
       "      <td>6.619658</td>\n",
       "      <td>6.692328</td>\n",
       "      <td>6.622764</td>\n",
       "      <td>...</td>\n",
       "      <td>6.558407</td>\n",
       "      <td>6.635161</td>\n",
       "      <td>6.541176</td>\n",
       "      <td>6.612361</td>\n",
       "      <td>6.504983</td>\n",
       "      <td>6.669707</td>\n",
       "      <td>6.513736</td>\n",
       "      <td>6.523818</td>\n",
       "      <td>6.742682</td>\n",
       "      <td>6.671303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311170</th>\n",
       "      <td>6.612094</td>\n",
       "      <td>6.658218</td>\n",
       "      <td>6.684843</td>\n",
       "      <td>6.737029</td>\n",
       "      <td>6.580584</td>\n",
       "      <td>6.616513</td>\n",
       "      <td>6.841786</td>\n",
       "      <td>6.555513</td>\n",
       "      <td>6.570710</td>\n",
       "      <td>6.673706</td>\n",
       "      <td>...</td>\n",
       "      <td>6.609160</td>\n",
       "      <td>6.714697</td>\n",
       "      <td>6.698058</td>\n",
       "      <td>6.727300</td>\n",
       "      <td>6.590967</td>\n",
       "      <td>6.690844</td>\n",
       "      <td>6.637789</td>\n",
       "      <td>6.635340</td>\n",
       "      <td>6.648415</td>\n",
       "      <td>6.549795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311175</th>\n",
       "      <td>6.650585</td>\n",
       "      <td>6.696714</td>\n",
       "      <td>6.680007</td>\n",
       "      <td>6.596399</td>\n",
       "      <td>6.775372</td>\n",
       "      <td>6.645024</td>\n",
       "      <td>6.636948</td>\n",
       "      <td>6.580950</td>\n",
       "      <td>6.587754</td>\n",
       "      <td>6.664373</td>\n",
       "      <td>...</td>\n",
       "      <td>6.706110</td>\n",
       "      <td>6.575635</td>\n",
       "      <td>6.578014</td>\n",
       "      <td>6.625735</td>\n",
       "      <td>6.599182</td>\n",
       "      <td>6.613412</td>\n",
       "      <td>6.723767</td>\n",
       "      <td>6.537973</td>\n",
       "      <td>6.747436</td>\n",
       "      <td>6.655572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311180</th>\n",
       "      <td>6.800705</td>\n",
       "      <td>6.635228</td>\n",
       "      <td>6.669978</td>\n",
       "      <td>6.597619</td>\n",
       "      <td>6.735752</td>\n",
       "      <td>6.652822</td>\n",
       "      <td>6.465435</td>\n",
       "      <td>6.728747</td>\n",
       "      <td>6.661455</td>\n",
       "      <td>6.540017</td>\n",
       "      <td>...</td>\n",
       "      <td>6.731054</td>\n",
       "      <td>6.616311</td>\n",
       "      <td>6.580413</td>\n",
       "      <td>6.639311</td>\n",
       "      <td>6.811333</td>\n",
       "      <td>6.688858</td>\n",
       "      <td>6.663167</td>\n",
       "      <td>6.675401</td>\n",
       "      <td>6.566967</td>\n",
       "      <td>6.683958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311185</th>\n",
       "      <td>6.652600</td>\n",
       "      <td>6.572627</td>\n",
       "      <td>6.685865</td>\n",
       "      <td>6.585365</td>\n",
       "      <td>6.652637</td>\n",
       "      <td>6.620325</td>\n",
       "      <td>6.667806</td>\n",
       "      <td>6.574783</td>\n",
       "      <td>6.688020</td>\n",
       "      <td>6.610882</td>\n",
       "      <td>...</td>\n",
       "      <td>6.573938</td>\n",
       "      <td>6.594607</td>\n",
       "      <td>6.529716</td>\n",
       "      <td>6.703237</td>\n",
       "      <td>6.568071</td>\n",
       "      <td>6.655248</td>\n",
       "      <td>6.584248</td>\n",
       "      <td>6.615269</td>\n",
       "      <td>6.664984</td>\n",
       "      <td>6.592061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311190</th>\n",
       "      <td>6.891066</td>\n",
       "      <td>6.650255</td>\n",
       "      <td>6.755141</td>\n",
       "      <td>6.694660</td>\n",
       "      <td>6.768542</td>\n",
       "      <td>6.629772</td>\n",
       "      <td>6.710507</td>\n",
       "      <td>6.587555</td>\n",
       "      <td>6.579731</td>\n",
       "      <td>6.709085</td>\n",
       "      <td>...</td>\n",
       "      <td>6.666897</td>\n",
       "      <td>6.809926</td>\n",
       "      <td>6.708028</td>\n",
       "      <td>6.895549</td>\n",
       "      <td>6.603061</td>\n",
       "      <td>6.730972</td>\n",
       "      <td>6.748203</td>\n",
       "      <td>6.790516</td>\n",
       "      <td>6.687368</td>\n",
       "      <td>6.489318</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47153 rows Ã— 242 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM2690230  GSM2690231  GSM2690232  GSM2690233  GSM2690234  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   14.415838   14.420469   14.454755   14.548113   14.393246   \n",
       "ILMN_1343295   12.501028   11.415616   11.695755   11.923042   11.554511   \n",
       "ILMN_1651199    6.627739    6.607227    6.585315    6.676918    6.565506   \n",
       "ILMN_1651209    6.556000    7.053035    6.802224    6.954742    6.890331   \n",
       "ILMN_1651210    6.597989    6.552052    6.658103    6.666267    6.727378   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_3311170    6.612094    6.658218    6.684843    6.737029    6.580584   \n",
       "ILMN_3311175    6.650585    6.696714    6.680007    6.596399    6.775372   \n",
       "ILMN_3311180    6.800705    6.635228    6.669978    6.597619    6.735752   \n",
       "ILMN_3311185    6.652600    6.572627    6.685865    6.585365    6.652637   \n",
       "ILMN_3311190    6.891066    6.650255    6.755141    6.694660    6.768542   \n",
       "\n",
       "              GSM2690235  GSM2690236  GSM2690237  GSM2690238  GSM2690239  ...  \\\n",
       "ID                                                                        ...   \n",
       "ILMN_1343291   14.476773   14.536234   14.655704   14.495031   14.532721  ...   \n",
       "ILMN_1343295   11.474396   12.061302   11.915192   11.952768   11.313382  ...   \n",
       "ILMN_1651199    6.706464    6.494084    6.584417    6.575962    6.658423  ...   \n",
       "ILMN_1651209    7.070821    6.954012    6.929918    6.918314    6.734604  ...   \n",
       "ILMN_1651210    6.667242    6.653112    6.619658    6.692328    6.622764  ...   \n",
       "...                  ...         ...         ...         ...         ...  ...   \n",
       "ILMN_3311170    6.616513    6.841786    6.555513    6.570710    6.673706  ...   \n",
       "ILMN_3311175    6.645024    6.636948    6.580950    6.587754    6.664373  ...   \n",
       "ILMN_3311180    6.652822    6.465435    6.728747    6.661455    6.540017  ...   \n",
       "ILMN_3311185    6.620325    6.667806    6.574783    6.688020    6.610882  ...   \n",
       "ILMN_3311190    6.629772    6.710507    6.587555    6.579731    6.709085  ...   \n",
       "\n",
       "              GSM2690462  GSM2690463  GSM2690464  GSM2690465  GSM2690466  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   14.649794   14.700669   14.376907   14.617264   14.512447   \n",
       "ILMN_1343295   11.580915   10.424846   10.646008   11.864454   11.133300   \n",
       "ILMN_1651199    6.702769    6.486990    6.786289    6.622436    6.590972   \n",
       "ILMN_1651209    6.800404    7.103074    6.853516    6.867996    6.674861   \n",
       "ILMN_1651210    6.558407    6.635161    6.541176    6.612361    6.504983   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_3311170    6.609160    6.714697    6.698058    6.727300    6.590967   \n",
       "ILMN_3311175    6.706110    6.575635    6.578014    6.625735    6.599182   \n",
       "ILMN_3311180    6.731054    6.616311    6.580413    6.639311    6.811333   \n",
       "ILMN_3311185    6.573938    6.594607    6.529716    6.703237    6.568071   \n",
       "ILMN_3311190    6.666897    6.809926    6.708028    6.895549    6.603061   \n",
       "\n",
       "              GSM2690467  GSM2690468  GSM2690469  GSM2690470  GSM2690471  \n",
       "ID                                                                        \n",
       "ILMN_1343291   14.295207   14.482401   14.478883   14.605284   14.556862  \n",
       "ILMN_1343295   11.915684   11.448013   11.947450   11.322478   11.192065  \n",
       "ILMN_1651199    6.481921    6.641028    6.683143    6.478434    6.537222  \n",
       "ILMN_1651209    6.775554    6.766406    6.716863    6.870444    6.863519  \n",
       "ILMN_1651210    6.669707    6.513736    6.523818    6.742682    6.671303  \n",
       "...                  ...         ...         ...         ...         ...  \n",
       "ILMN_3311170    6.690844    6.637789    6.635340    6.648415    6.549795  \n",
       "ILMN_3311175    6.613412    6.723767    6.537973    6.747436    6.655572  \n",
       "ILMN_3311180    6.688858    6.663167    6.675401    6.566967    6.683958  \n",
       "ILMN_3311185    6.655248    6.584248    6.615269    6.664984    6.592061  \n",
       "ILMN_3311190    6.730972    6.748203    6.790516    6.687368    6.489318  \n",
       "\n",
       "[47153 rows x 242 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651199', 'ILMN_1651209', 'ILMN_1651210', 'ILMN_1651221', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651230', 'ILMN_1651232', 'ILMN_1651235', 'ILMN_1651236', 'ILMN_1651237', 'ILMN_1651238', 'ILMN_1651253', 'ILMN_1651254', 'ILMN_1651259', 'ILMN_1651260', 'ILMN_1651262', 'ILMN_1651268']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Source</th>\n",
       "      <th>Search_Key</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>ILMN_Gene</th>\n",
       "      <th>Source_Reference_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>Unigene_ID</th>\n",
       "      <th>Entrez_Gene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Probe_Chr_Orientation</th>\n",
       "      <th>Probe_Coordinates</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Ontology_Component</th>\n",
       "      <th>Ontology_Process</th>\n",
       "      <th>Ontology_Function</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Obsolete_Probe_Id</th>\n",
       "      <th>GB_ACC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ILMN_1343048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ILMN_1343049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ILMN_1343050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ILMN_1343052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ILMN_1343059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11459370</th>\n",
       "      <td>ILMN_2371169</td>\n",
       "      <td>7.855694502</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11459371</th>\n",
       "      <td>ILMN_1701875</td>\n",
       "      <td>8.956319227</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11459372</th>\n",
       "      <td>ILMN_1786396</td>\n",
       "      <td>8.314829409</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11459373</th>\n",
       "      <td>ILMN_1653618</td>\n",
       "      <td>8.968126225</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11459374</th>\n",
       "      <td>ILMN_2137536</td>\n",
       "      <td>8.777659106</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11459375 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID      Species Source Search_Key Transcript ILMN_Gene  \\\n",
       "0         ILMN_1343048          NaN    NaN        NaN        NaN       NaN   \n",
       "1         ILMN_1343049          NaN    NaN        NaN        NaN       NaN   \n",
       "2         ILMN_1343050          NaN    NaN        NaN        NaN       NaN   \n",
       "3         ILMN_1343052          NaN    NaN        NaN        NaN       NaN   \n",
       "4         ILMN_1343059          NaN    NaN        NaN        NaN       NaN   \n",
       "...                ...          ...    ...        ...        ...       ...   \n",
       "11459370  ILMN_2371169  7.855694502      0        NaN        NaN       NaN   \n",
       "11459371  ILMN_1701875  8.956319227      0        NaN        NaN       NaN   \n",
       "11459372  ILMN_1786396  8.314829409      0        NaN        NaN       NaN   \n",
       "11459373  ILMN_1653618  8.968126225      0        NaN        NaN       NaN   \n",
       "11459374  ILMN_2137536  8.777659106      0        NaN        NaN       NaN   \n",
       "\n",
       "         Source_Reference_ID RefSeq_ID Unigene_ID  Entrez_Gene_ID  ...  \\\n",
       "0                        NaN       NaN        NaN             NaN  ...   \n",
       "1                        NaN       NaN        NaN             NaN  ...   \n",
       "2                        NaN       NaN        NaN             NaN  ...   \n",
       "3                        NaN       NaN        NaN             NaN  ...   \n",
       "4                        NaN       NaN        NaN             NaN  ...   \n",
       "...                      ...       ...        ...             ...  ...   \n",
       "11459370                 NaN       NaN        NaN             NaN  ...   \n",
       "11459371                 NaN       NaN        NaN             NaN  ...   \n",
       "11459372                 NaN       NaN        NaN             NaN  ...   \n",
       "11459373                 NaN       NaN        NaN             NaN  ...   \n",
       "11459374                 NaN       NaN        NaN             NaN  ...   \n",
       "\n",
       "          Probe_Chr_Orientation Probe_Coordinates Cytoband Definition  \\\n",
       "0                           NaN               NaN      NaN        NaN   \n",
       "1                           NaN               NaN      NaN        NaN   \n",
       "2                           NaN               NaN      NaN        NaN   \n",
       "3                           NaN               NaN      NaN        NaN   \n",
       "4                           NaN               NaN      NaN        NaN   \n",
       "...                         ...               ...      ...        ...   \n",
       "11459370                    NaN               NaN      NaN        NaN   \n",
       "11459371                    NaN               NaN      NaN        NaN   \n",
       "11459372                    NaN               NaN      NaN        NaN   \n",
       "11459373                    NaN               NaN      NaN        NaN   \n",
       "11459374                    NaN               NaN      NaN        NaN   \n",
       "\n",
       "         Ontology_Component  Ontology_Process Ontology_Function  Synonyms  \\\n",
       "0                       NaN               NaN               NaN       NaN   \n",
       "1                       NaN               NaN               NaN       NaN   \n",
       "2                       NaN               NaN               NaN       NaN   \n",
       "3                       NaN               NaN               NaN       NaN   \n",
       "4                       NaN               NaN               NaN       NaN   \n",
       "...                     ...               ...               ...       ...   \n",
       "11459370                NaN               NaN               NaN       NaN   \n",
       "11459371                NaN               NaN               NaN       NaN   \n",
       "11459372                NaN               NaN               NaN       NaN   \n",
       "11459373                NaN               NaN               NaN       NaN   \n",
       "11459374                NaN               NaN               NaN       NaN   \n",
       "\n",
       "         Obsolete_Probe_Id GB_ACC  \n",
       "0                      NaN    NaN  \n",
       "1                      NaN    NaN  \n",
       "2                      NaN    NaN  \n",
       "3                      NaN    NaN  \n",
       "4                      NaN    NaN  \n",
       "...                    ...    ...  \n",
       "11459370               NaN    NaN  \n",
       "11459371               NaN    NaN  \n",
       "11459372               NaN    NaN  \n",
       "11459373               NaN    NaN  \n",
       "11459374               NaN    NaN  \n",
       "\n",
       "[11459375 rows x 30 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Species', 'Source', 'Search_Key', 'Transcript', 'ILMN_Gene',\n",
       "       'Source_Reference_ID', 'RefSeq_ID', 'Unigene_ID', 'Entrez_Gene_ID',\n",
       "       'GI', 'Accession', 'Symbol', 'Protein_Product', 'Probe_Id',\n",
       "       'Array_Address_Id', 'Probe_Type', 'Probe_Start', 'SEQUENCE',\n",
       "       'Chromosome', 'Probe_Chr_Orientation', 'Probe_Coordinates', 'Cytoband',\n",
       "       'Definition', 'Ontology_Component', 'Ontology_Process',\n",
       "       'Ontology_Function', 'Synonyms', 'Obsolete_Probe_Id', 'GB_ACC'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2690230</th>\n",
       "      <th>GSM2690231</th>\n",
       "      <th>GSM2690232</th>\n",
       "      <th>GSM2690233</th>\n",
       "      <th>GSM2690234</th>\n",
       "      <th>GSM2690235</th>\n",
       "      <th>GSM2690236</th>\n",
       "      <th>GSM2690237</th>\n",
       "      <th>GSM2690238</th>\n",
       "      <th>GSM2690239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2690462</th>\n",
       "      <th>GSM2690463</th>\n",
       "      <th>GSM2690464</th>\n",
       "      <th>GSM2690465</th>\n",
       "      <th>GSM2690466</th>\n",
       "      <th>GSM2690467</th>\n",
       "      <th>GSM2690468</th>\n",
       "      <th>GSM2690469</th>\n",
       "      <th>GSM2690470</th>\n",
       "      <th>GSM2690471</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1-Dec</th>\n",
       "      <td>6.690225</td>\n",
       "      <td>6.645631</td>\n",
       "      <td>6.613515</td>\n",
       "      <td>6.639225</td>\n",
       "      <td>6.705352</td>\n",
       "      <td>6.642645</td>\n",
       "      <td>6.563353</td>\n",
       "      <td>6.563837</td>\n",
       "      <td>6.530434</td>\n",
       "      <td>6.581223</td>\n",
       "      <td>...</td>\n",
       "      <td>6.634416</td>\n",
       "      <td>6.510920</td>\n",
       "      <td>6.563834</td>\n",
       "      <td>6.627330</td>\n",
       "      <td>6.585572</td>\n",
       "      <td>6.701072</td>\n",
       "      <td>6.556956</td>\n",
       "      <td>6.683356</td>\n",
       "      <td>6.609795</td>\n",
       "      <td>6.650483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-Mar</th>\n",
       "      <td>7.651063</td>\n",
       "      <td>9.571072</td>\n",
       "      <td>8.556212</td>\n",
       "      <td>8.587053</td>\n",
       "      <td>9.433715</td>\n",
       "      <td>9.594437</td>\n",
       "      <td>9.409010</td>\n",
       "      <td>9.165061</td>\n",
       "      <td>9.791250</td>\n",
       "      <td>9.178682</td>\n",
       "      <td>...</td>\n",
       "      <td>8.791651</td>\n",
       "      <td>9.323251</td>\n",
       "      <td>9.518603</td>\n",
       "      <td>8.650389</td>\n",
       "      <td>9.074401</td>\n",
       "      <td>9.056935</td>\n",
       "      <td>9.216715</td>\n",
       "      <td>8.608755</td>\n",
       "      <td>9.224840</td>\n",
       "      <td>8.558667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10-Mar</th>\n",
       "      <td>6.826650</td>\n",
       "      <td>6.588074</td>\n",
       "      <td>6.590686</td>\n",
       "      <td>6.597999</td>\n",
       "      <td>6.646043</td>\n",
       "      <td>6.676378</td>\n",
       "      <td>6.596264</td>\n",
       "      <td>6.579935</td>\n",
       "      <td>6.616365</td>\n",
       "      <td>6.645453</td>\n",
       "      <td>...</td>\n",
       "      <td>6.559374</td>\n",
       "      <td>6.610369</td>\n",
       "      <td>6.582099</td>\n",
       "      <td>6.742123</td>\n",
       "      <td>6.607611</td>\n",
       "      <td>6.592589</td>\n",
       "      <td>6.641961</td>\n",
       "      <td>6.484952</td>\n",
       "      <td>6.527867</td>\n",
       "      <td>6.685483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11-Mar</th>\n",
       "      <td>6.414740</td>\n",
       "      <td>6.579584</td>\n",
       "      <td>6.574483</td>\n",
       "      <td>6.563864</td>\n",
       "      <td>6.561773</td>\n",
       "      <td>6.560585</td>\n",
       "      <td>6.485683</td>\n",
       "      <td>6.580335</td>\n",
       "      <td>6.443181</td>\n",
       "      <td>6.738039</td>\n",
       "      <td>...</td>\n",
       "      <td>6.591454</td>\n",
       "      <td>6.522802</td>\n",
       "      <td>6.542323</td>\n",
       "      <td>6.517765</td>\n",
       "      <td>6.697750</td>\n",
       "      <td>6.657025</td>\n",
       "      <td>6.722615</td>\n",
       "      <td>6.793510</td>\n",
       "      <td>6.578635</td>\n",
       "      <td>6.436090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-Mar</th>\n",
       "      <td>7.620253</td>\n",
       "      <td>7.269074</td>\n",
       "      <td>7.263967</td>\n",
       "      <td>7.332735</td>\n",
       "      <td>7.215201</td>\n",
       "      <td>7.068063</td>\n",
       "      <td>7.490501</td>\n",
       "      <td>7.255800</td>\n",
       "      <td>7.211440</td>\n",
       "      <td>7.370995</td>\n",
       "      <td>...</td>\n",
       "      <td>7.335041</td>\n",
       "      <td>7.171274</td>\n",
       "      <td>7.036233</td>\n",
       "      <td>7.365702</td>\n",
       "      <td>7.365079</td>\n",
       "      <td>7.148670</td>\n",
       "      <td>7.330081</td>\n",
       "      <td>7.514638</td>\n",
       "      <td>7.421067</td>\n",
       "      <td>7.287512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <td>7.872240</td>\n",
       "      <td>7.808094</td>\n",
       "      <td>7.804654</td>\n",
       "      <td>7.844383</td>\n",
       "      <td>7.853817</td>\n",
       "      <td>7.792229</td>\n",
       "      <td>8.296184</td>\n",
       "      <td>8.005402</td>\n",
       "      <td>7.756242</td>\n",
       "      <td>8.175150</td>\n",
       "      <td>...</td>\n",
       "      <td>8.009629</td>\n",
       "      <td>7.702242</td>\n",
       "      <td>7.744495</td>\n",
       "      <td>7.510275</td>\n",
       "      <td>7.649329</td>\n",
       "      <td>7.759432</td>\n",
       "      <td>7.742089</td>\n",
       "      <td>7.757284</td>\n",
       "      <td>7.602962</td>\n",
       "      <td>7.589611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gm127</th>\n",
       "      <td>6.652717</td>\n",
       "      <td>6.661334</td>\n",
       "      <td>6.766349</td>\n",
       "      <td>6.766286</td>\n",
       "      <td>6.726882</td>\n",
       "      <td>6.739457</td>\n",
       "      <td>6.691700</td>\n",
       "      <td>6.836504</td>\n",
       "      <td>6.715168</td>\n",
       "      <td>6.611255</td>\n",
       "      <td>...</td>\n",
       "      <td>6.781494</td>\n",
       "      <td>6.657049</td>\n",
       "      <td>6.701411</td>\n",
       "      <td>6.758011</td>\n",
       "      <td>6.802065</td>\n",
       "      <td>6.699023</td>\n",
       "      <td>6.653731</td>\n",
       "      <td>6.795169</td>\n",
       "      <td>6.829674</td>\n",
       "      <td>6.672595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psiTPTE22</th>\n",
       "      <td>6.543646</td>\n",
       "      <td>6.651050</td>\n",
       "      <td>6.592777</td>\n",
       "      <td>6.712322</td>\n",
       "      <td>6.651795</td>\n",
       "      <td>6.687181</td>\n",
       "      <td>6.722882</td>\n",
       "      <td>6.615713</td>\n",
       "      <td>6.647023</td>\n",
       "      <td>6.651957</td>\n",
       "      <td>...</td>\n",
       "      <td>6.658668</td>\n",
       "      <td>6.648473</td>\n",
       "      <td>6.719836</td>\n",
       "      <td>6.719294</td>\n",
       "      <td>6.723473</td>\n",
       "      <td>6.659385</td>\n",
       "      <td>6.640942</td>\n",
       "      <td>6.653864</td>\n",
       "      <td>6.602776</td>\n",
       "      <td>6.540232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rab1c</th>\n",
       "      <td>6.720996</td>\n",
       "      <td>6.614096</td>\n",
       "      <td>6.592334</td>\n",
       "      <td>6.602132</td>\n",
       "      <td>6.775020</td>\n",
       "      <td>6.598250</td>\n",
       "      <td>6.731486</td>\n",
       "      <td>6.685775</td>\n",
       "      <td>6.855869</td>\n",
       "      <td>6.709597</td>\n",
       "      <td>...</td>\n",
       "      <td>6.507397</td>\n",
       "      <td>6.568342</td>\n",
       "      <td>6.536055</td>\n",
       "      <td>6.523492</td>\n",
       "      <td>6.719801</td>\n",
       "      <td>6.596157</td>\n",
       "      <td>6.563001</td>\n",
       "      <td>6.661075</td>\n",
       "      <td>6.639940</td>\n",
       "      <td>6.565698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tAKR</th>\n",
       "      <td>6.709635</td>\n",
       "      <td>6.586229</td>\n",
       "      <td>6.532554</td>\n",
       "      <td>6.620359</td>\n",
       "      <td>6.677593</td>\n",
       "      <td>6.596459</td>\n",
       "      <td>6.612589</td>\n",
       "      <td>6.630510</td>\n",
       "      <td>6.539535</td>\n",
       "      <td>6.658835</td>\n",
       "      <td>...</td>\n",
       "      <td>6.679438</td>\n",
       "      <td>6.583316</td>\n",
       "      <td>6.624987</td>\n",
       "      <td>6.620980</td>\n",
       "      <td>6.562871</td>\n",
       "      <td>6.660996</td>\n",
       "      <td>6.602719</td>\n",
       "      <td>6.503069</td>\n",
       "      <td>6.625130</td>\n",
       "      <td>6.723288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31335 rows Ã— 242 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM2690230  GSM2690231  GSM2690232  GSM2690233  GSM2690234  \\\n",
       "Gene                                                                     \n",
       "1-Dec         6.690225    6.645631    6.613515    6.639225    6.705352   \n",
       "1-Mar         7.651063    9.571072    8.556212    8.587053    9.433715   \n",
       "10-Mar        6.826650    6.588074    6.590686    6.597999    6.646043   \n",
       "11-Mar        6.414740    6.579584    6.574483    6.563864    6.561773   \n",
       "2-Mar         7.620253    7.269074    7.263967    7.332735    7.215201   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "dJ341D10.1    7.872240    7.808094    7.804654    7.844383    7.853817   \n",
       "gm127         6.652717    6.661334    6.766349    6.766286    6.726882   \n",
       "psiTPTE22     6.543646    6.651050    6.592777    6.712322    6.651795   \n",
       "rab1c         6.720996    6.614096    6.592334    6.602132    6.775020   \n",
       "tAKR          6.709635    6.586229    6.532554    6.620359    6.677593   \n",
       "\n",
       "            GSM2690235  GSM2690236  GSM2690237  GSM2690238  GSM2690239  ...  \\\n",
       "Gene                                                                    ...   \n",
       "1-Dec         6.642645    6.563353    6.563837    6.530434    6.581223  ...   \n",
       "1-Mar         9.594437    9.409010    9.165061    9.791250    9.178682  ...   \n",
       "10-Mar        6.676378    6.596264    6.579935    6.616365    6.645453  ...   \n",
       "11-Mar        6.560585    6.485683    6.580335    6.443181    6.738039  ...   \n",
       "2-Mar         7.068063    7.490501    7.255800    7.211440    7.370995  ...   \n",
       "...                ...         ...         ...         ...         ...  ...   \n",
       "dJ341D10.1    7.792229    8.296184    8.005402    7.756242    8.175150  ...   \n",
       "gm127         6.739457    6.691700    6.836504    6.715168    6.611255  ...   \n",
       "psiTPTE22     6.687181    6.722882    6.615713    6.647023    6.651957  ...   \n",
       "rab1c         6.598250    6.731486    6.685775    6.855869    6.709597  ...   \n",
       "tAKR          6.596459    6.612589    6.630510    6.539535    6.658835  ...   \n",
       "\n",
       "            GSM2690462  GSM2690463  GSM2690464  GSM2690465  GSM2690466  \\\n",
       "Gene                                                                     \n",
       "1-Dec         6.634416    6.510920    6.563834    6.627330    6.585572   \n",
       "1-Mar         8.791651    9.323251    9.518603    8.650389    9.074401   \n",
       "10-Mar        6.559374    6.610369    6.582099    6.742123    6.607611   \n",
       "11-Mar        6.591454    6.522802    6.542323    6.517765    6.697750   \n",
       "2-Mar         7.335041    7.171274    7.036233    7.365702    7.365079   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "dJ341D10.1    8.009629    7.702242    7.744495    7.510275    7.649329   \n",
       "gm127         6.781494    6.657049    6.701411    6.758011    6.802065   \n",
       "psiTPTE22     6.658668    6.648473    6.719836    6.719294    6.723473   \n",
       "rab1c         6.507397    6.568342    6.536055    6.523492    6.719801   \n",
       "tAKR          6.679438    6.583316    6.624987    6.620980    6.562871   \n",
       "\n",
       "            GSM2690467  GSM2690468  GSM2690469  GSM2690470  GSM2690471  \n",
       "Gene                                                                    \n",
       "1-Dec         6.701072    6.556956    6.683356    6.609795    6.650483  \n",
       "1-Mar         9.056935    9.216715    8.608755    9.224840    8.558667  \n",
       "10-Mar        6.592589    6.641961    6.484952    6.527867    6.685483  \n",
       "11-Mar        6.657025    6.722615    6.793510    6.578635    6.436090  \n",
       "2-Mar         7.148670    7.330081    7.514638    7.421067    7.287512  \n",
       "...                ...         ...         ...         ...         ...  \n",
       "dJ341D10.1    7.759432    7.742089    7.757284    7.602962    7.589611  \n",
       "gm127         6.699023    6.653731    6.795169    6.829674    6.672595  \n",
       "psiTPTE22     6.659385    6.640942    6.653864    6.602776    6.540232  \n",
       "rab1c         6.596157    6.563001    6.661075    6.639940    6.565698  \n",
       "tAKR          6.660996    6.602719    6.503069    6.625130    6.723288  \n",
       "\n",
       "[31335 rows x 242 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13 input query terms found dup hits:\t[('ABCC13', 2), ('ABCC6P1', 2), ('ABCC6P2', 3), ('ADAM3A', 2), ('ADAM6', 3), ('AGAP11', 2), ('AK2P2'\n",
      "106 input query terms found no hit:\t['1-Dec', '1-Mar', '10-Mar', '11-Mar', '2-Mar', '3-Mar', '4-Mar', '5-Mar', '6-Mar', '7-Mar', '7A5', \n",
      "6 input query terms found dup hits:\t[('ATP6AP1L', 2), ('ATXN8OS', 2), ('BAGE2', 2), ('BIRC8', 2), ('BRD7P3', 2), ('BRI3P1', 2)]\n",
      "363 input query terms found no hit:\t['ARMC4', 'ARMET', 'ARNTL', 'ARNTL2', 'ARP11', 'ARPM1', 'ARPM2', 'ARPP-21', 'ARS2', 'ARSE', 'ARVP612\n",
      "809 input query terms found no hit:\t['C15orf42', 'C15orf43', 'C15orf44', 'C15orf49', 'C15orf5', 'C15orf51', 'C15orf52', 'C15orf53', 'C15\n",
      "7 input query terms found dup hits:\t[('CATSPER2P1', 2), ('CCDC144NL', 2), ('CCNYL2', 2), ('CCNYL3', 2), ('CCT6P1', 2), ('CDR1', 2), ('CE\n",
      "131 input query terms found no hit:\t['C9orf90', 'C9orf91', 'C9orf93', 'C9orf96', 'C9orf98', 'CA5BP', 'CABC1', 'CAMSAP1L1', 'CARD17', 'CA\n",
      "4 input query terms found dup hits:\t[('CLEC4GP1', 2), ('COPG2IT1', 2), ('CXADRP2', 3), ('CYP4Z2P', 2)]\n",
      "119 input query terms found no hit:\t['CLECL1', 'CLK2P', 'CLLU1OS', 'CLRN1OS', 'CMAH', 'CMYA3', 'CNIH', 'CNO', 'CNTD2', 'COBRA1', 'COL28'\n",
      "12 input query terms found dup hits:\t[('DEFB117', 2), ('DEFB122', 2), ('DHRS4L1', 3), ('DIRC1', 2), ('DLEU2L', 2), ('DNAJB3', 2), ('DNM1P\n",
      "209 input query terms found no hit:\t['DEFB130', 'DEFB131', 'DEFB137', 'DEFB32', 'DEFB4', 'DEM1', 'DEPDC6', 'DFNA5', 'DFNB31', 'DFNB59', \n",
      "9 input query terms found dup hits:\t[('ESPNP', 2), ('ESRRAP2', 2), ('FAM197Y2', 3), ('FAM41AY2', 2), ('FAM74A3', 2), ('FAM74A4', 2), ('F\n",
      "427 input query terms found no hit:\t['ERCC-00162', 'ERCC-00163', 'ERCC-00164', 'ERCC-00165', 'ERCC-00168', 'ERCC-00170', 'ERCC-00171', '\n",
      "10 input query terms found dup hits:\t[('FMO6P', 2), ('FMO9P', 2), ('FOLH1B', 2), ('GABPAP', 2), ('GAGE12B', 2), ('GGNBP1', 2), ('GGT8P', \n",
      "288 input query terms found no hit:\t['FLJ35785', 'FLJ35801', 'FLJ35816', 'FLJ35848', 'FLJ35880', 'FLJ35894', 'FLJ35934', 'FLJ35946', 'FL\n",
      "18 input query terms found dup hits:\t[('GTF2H2B', 2), ('GTF2IP1', 2), ('GUCY1B2', 2), ('HAR1A', 2), ('HBBP1', 2), ('HBD', 2), ('HCG4', 9)\n",
      "208 input query terms found no hit:\t['GPR158L1', 'GPR172A', 'GPR172B', 'GPR175', 'GPR177', 'GPR44', 'GPR56', 'GPR64', 'GPR77', 'GPR81', \n",
      "3 input query terms found dup hits:\t[('INGX', 2), ('KIR3DX1', 5), ('KLKP1', 2)]\n",
      "225 input query terms found no hit:\t['IFNE1', 'IFP38', 'IFRG15', 'IGF2AS', 'IGJ', 'IGLL3', 'IGSF2', 'IHPK3', 'IIP45', 'IKBKAP', 'IL1F5',\n",
      "3 input query terms found dup hits:\t[('KRT126P', 2), ('KRT18P19', 2), ('KRT18P40', 3)]\n",
      "535 input query terms found no hit:\t['KRTAP9-5', 'KRTHA5', 'KRTHB6', 'KTELC1', 'L3MBTL', 'LACE1', 'LAMB2L', 'LARGE', 'LARS', 'LASS1', 'L\n",
      "955 input query terms found no hit:\t['LOC100129660', 'LOC100129663', 'LOC100129665', 'LOC100129668', 'LOC100129673', 'LOC100129674', 'LO\n",
      "972 input query terms found no hit:\t['LOC100132869', 'LOC100132870', 'LOC100132878', 'LOC100132880', 'LOC100132883', 'LOC100132884', 'LO\n",
      "923 input query terms found no hit:\t['LOC283487', 'LOC283514', 'LOC283523', 'LOC283547', 'LOC283553', 'LOC283584', 'LOC283585', 'LOC2835\n",
      "945 input query terms found no hit:\t['LOC440225', 'LOC440226', 'LOC440233', 'LOC440243', 'LOC440248', 'LOC440258', 'LOC440261', 'LOC4402\n",
      "967 input query terms found no hit:\t['LOC643152', 'LOC643153', 'LOC643154', 'LOC643158', 'LOC643159', 'LOC643161', 'LOC643167', 'LOC6431\n",
      "960 input query terms found no hit:\t['LOC645362', 'LOC645363', 'LOC645364', 'LOC645365', 'LOC645367', 'LOC645373', 'LOC645378', 'LOC6453\n",
      "994 input query terms found no hit:\t['LOC648111', 'LOC648117', 'LOC648123', 'LOC648127', 'LOC648130', 'LOC648132', 'LOC648133', 'LOC6481\n",
      "991 input query terms found no hit:\t['LOC651769', 'LOC651771', 'LOC651772', 'LOC651774', 'LOC651777', 'LOC651787', 'LOC651789', 'LOC6517\n",
      "944 input query terms found no hit:\t['LOC727768', 'LOC727773', 'LOC727775', 'LOC727778', 'LOC727781', 'LOC727787', 'LOC727788', 'LOC7277\n",
      "6 input query terms found dup hits:\t[('LPAL2', 2), ('LRP5L', 2), ('LY6G6E', 8), ('LYPLA2P1', 6), ('MEG8', 2), ('MEIS3P1', 2)]\n",
      "364 input query terms found no hit:\t['LOC730474', 'LOC730476', 'LOC730479', 'LOC730496', 'LOC730510', 'LOC730517', 'LOC730525', 'LOC7305\n",
      "9 input query terms found dup hits:\t[('MIR320B2', 2), ('MMD2', 2), ('MORF4', 2), ('MRS2P2', 2), ('MSTO2P', 2), ('MSX2P1', 2), ('MT1IP', \n",
      "124 input query terms found no hit:\t['MGC3771', 'MGC39372', 'MGC39545', 'MGC39581', 'MGC39584', 'MGC39606', 'MGC39821', 'MGC39900', 'MGC\n",
      "11 input query terms found dup hits:\t[('MYADML', 2), ('NAPSB', 2), ('NAT8B', 3), ('NBPF22P', 2), ('NCF1B', 2), ('NME2P1', 2), ('NPY6R', 2\n",
      "125 input query terms found no hit:\t['MTE', 'MTERF', 'MTERFD1', 'MTERFD2', 'MTERFD3', 'MTL5', 'MTM', 'MTMR15', 'MTP18', 'MTSS1L', 'MUCDH\n",
      "11 input query terms found dup hits:\t[('OR10J3', 3), ('OR2A9P', 2), ('OR6W1P', 3), ('OR7E156P', 2), ('OR7E37P', 2), ('OR7E91P', 2), ('P2R\n",
      "64 input query terms found no hit:\t['OKL38', 'OPRS1', 'OR2B3P', 'OR2W5', 'OR3A4', 'OR52A4', 'OR5AY1', 'OR5R1', 'OR8G2', 'ORAOV1', 'ORC1\n",
      "13 input query terms found dup hits:\t[('PGM5P2', 2), ('PI4KAP1', 2), ('PKD1P1', 2), ('PLGLA', 2), ('PMCHL1', 2), ('PNMA6B', 2), ('POM121L\n",
      "104 input query terms found no hit:\t['PGCP', 'PH-4', 'PHACS', 'PHB', 'PHCA', 'PHF15', 'PHF16', 'PHF17', 'PIB5PA', 'PIK4CA', 'PIN1L', 'PI\n",
      "25 input query terms found dup hits:\t[('PSORS1C3', 8), ('PTENP1', 2), ('PVT1', 2), ('PYY2', 2), ('RAET1K', 2), ('RAP1BL', 2), ('RBMY1A3P'\n",
      "93 input query terms found no hit:\t['PSPHL', 'PTCHD2', 'PTK9', 'PTPLA', 'PTPLAD1', 'PTPLAD2', 'PTPLB', 'PTPN20A', 'PTPN20B', 'PTPRV', '\n",
      "13 input query terms found dup hits:\t[('RPLP0P2', 2), ('RPPH1', 2), ('RPS26P11', 2), ('S100A7L2', 2), ('SCARNA17', 2), ('SCARNA4', 2), ('\n",
      "100 input query terms found no hit:\t['RPRC1', 'RPS26L', 'RQCD1', 'RRP7B', 'RSHL1', 'RSHL3', 'RTCD1', 'RTDR1', 'RUNDC2A', 'RUNDC2B', 'RUN\n",
      "30 input query terms found dup hits:\t[('SLC6A10P', 2), ('SNHG6', 2), ('SNHG8', 2), ('SNHG9', 3), ('SNORA16A', 2), ('SNORA58', 2), ('SNORA\n",
      "59 input query terms found no hit:\t['SLC35E2', 'SLC9A10', 'SLC9A11', 'SLC9A3R1', 'SLC9A3R2', 'SLIC1', 'SLMO1', 'SLMO2', 'SMA3', 'SMA4',\n",
      "11 input query terms found dup hits:\t[('STAG3L1', 2), ('STAG3L2', 2), ('STAG3L3', 2), ('STAG3L4', 2), ('SUMO1P1', 2), ('SUMO1P3', 2), ('T\n",
      "120 input query terms found no hit:\t['SSX6', 'SSX8', 'SSX9', 'ST5', 'ST7OT1', 'ST7OT2', 'ST7OT3', 'ST7OT4', 'STELLAR', 'STGC3', 'STRA13'\n",
      "13 input query terms found dup hits:\t[('TNXA', 5), ('TOP1P1', 2), ('TPRXL', 2), ('TRIM16L', 2), ('TRPC2', 2), ('TTTY11', 2), ('TTTY17C', \n",
      "106 input query terms found no hit:\t['TMEM206', 'TMEM211', 'TMEM22', 'TMEM27', 'TMEM34', 'TMEM35', 'TMEM4', 'TMEM48', 'TMEM49', 'TMEM5',\n",
      "7 input query terms found dup hits:\t[('VENTXP1', 2), ('VENTXP7', 2), ('WASH2P', 2), ('WASH3P', 3), ('WASH5P', 2), ('WBP11P1', 2), ('XKRY\n",
      "124 input query terms found no hit:\t['USH3A', 'USMG5', 'USP17', 'USP41', 'UTP11L', 'UTS2D', 'UTX', 'VARS', 'VCX-C', 'VIL2', 'VISA', 'VMD\n",
      "4 input query terms found dup hits:\t[('ZNF658B', 2), ('ZNF702P', 2), ('ZNF781', 2), ('rab1c', 2)]\n",
      "30 input query terms found no hit:\t['ZNF509', 'ZNF533', 'ZNF542', 'ZNF642', 'ZNF643', 'ZNF645', 'ZNF650', 'ZNF663', 'ZNF673', 'ZNF720',\n"
     ]
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2690230</th>\n",
       "      <th>GSM2690231</th>\n",
       "      <th>GSM2690232</th>\n",
       "      <th>GSM2690233</th>\n",
       "      <th>GSM2690234</th>\n",
       "      <th>GSM2690235</th>\n",
       "      <th>GSM2690236</th>\n",
       "      <th>GSM2690237</th>\n",
       "      <th>GSM2690238</th>\n",
       "      <th>GSM2690239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2690462</th>\n",
       "      <th>GSM2690463</th>\n",
       "      <th>GSM2690464</th>\n",
       "      <th>GSM2690465</th>\n",
       "      <th>GSM2690466</th>\n",
       "      <th>GSM2690467</th>\n",
       "      <th>GSM2690468</th>\n",
       "      <th>GSM2690469</th>\n",
       "      <th>GSM2690470</th>\n",
       "      <th>GSM2690471</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>6.848050</td>\n",
       "      <td>6.688162</td>\n",
       "      <td>6.796304</td>\n",
       "      <td>6.726489</td>\n",
       "      <td>6.786394</td>\n",
       "      <td>6.854404</td>\n",
       "      <td>6.674512</td>\n",
       "      <td>6.688926</td>\n",
       "      <td>6.807333</td>\n",
       "      <td>6.666278</td>\n",
       "      <td>...</td>\n",
       "      <td>6.783802</td>\n",
       "      <td>6.657357</td>\n",
       "      <td>6.717128</td>\n",
       "      <td>6.735756</td>\n",
       "      <td>6.748284</td>\n",
       "      <td>6.761963</td>\n",
       "      <td>6.747374</td>\n",
       "      <td>6.804050</td>\n",
       "      <td>6.778584</td>\n",
       "      <td>6.751946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>6.583041</td>\n",
       "      <td>6.650677</td>\n",
       "      <td>6.686538</td>\n",
       "      <td>6.677932</td>\n",
       "      <td>6.661399</td>\n",
       "      <td>6.609482</td>\n",
       "      <td>6.628126</td>\n",
       "      <td>6.613500</td>\n",
       "      <td>6.613940</td>\n",
       "      <td>6.596914</td>\n",
       "      <td>...</td>\n",
       "      <td>6.574616</td>\n",
       "      <td>6.561873</td>\n",
       "      <td>6.650160</td>\n",
       "      <td>6.606254</td>\n",
       "      <td>6.663059</td>\n",
       "      <td>6.624456</td>\n",
       "      <td>6.657874</td>\n",
       "      <td>6.672318</td>\n",
       "      <td>6.741131</td>\n",
       "      <td>6.641459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>6.650136</td>\n",
       "      <td>6.521780</td>\n",
       "      <td>6.566705</td>\n",
       "      <td>6.618576</td>\n",
       "      <td>6.580874</td>\n",
       "      <td>6.569995</td>\n",
       "      <td>6.588907</td>\n",
       "      <td>6.559197</td>\n",
       "      <td>6.623999</td>\n",
       "      <td>6.567076</td>\n",
       "      <td>...</td>\n",
       "      <td>6.633357</td>\n",
       "      <td>6.453508</td>\n",
       "      <td>6.538286</td>\n",
       "      <td>6.437383</td>\n",
       "      <td>6.601649</td>\n",
       "      <td>6.726172</td>\n",
       "      <td>6.425379</td>\n",
       "      <td>6.620081</td>\n",
       "      <td>6.528318</td>\n",
       "      <td>6.530385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>6.611510</td>\n",
       "      <td>6.646145</td>\n",
       "      <td>6.633586</td>\n",
       "      <td>6.460136</td>\n",
       "      <td>6.465241</td>\n",
       "      <td>6.422501</td>\n",
       "      <td>6.467049</td>\n",
       "      <td>6.607412</td>\n",
       "      <td>6.529374</td>\n",
       "      <td>6.381183</td>\n",
       "      <td>...</td>\n",
       "      <td>6.454338</td>\n",
       "      <td>6.698004</td>\n",
       "      <td>6.722807</td>\n",
       "      <td>6.648967</td>\n",
       "      <td>6.501302</td>\n",
       "      <td>6.586460</td>\n",
       "      <td>6.566908</td>\n",
       "      <td>6.641582</td>\n",
       "      <td>6.606379</td>\n",
       "      <td>6.488792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>6.527584</td>\n",
       "      <td>6.622752</td>\n",
       "      <td>6.591259</td>\n",
       "      <td>6.624255</td>\n",
       "      <td>6.512508</td>\n",
       "      <td>6.553509</td>\n",
       "      <td>6.691260</td>\n",
       "      <td>6.716285</td>\n",
       "      <td>6.605287</td>\n",
       "      <td>6.574475</td>\n",
       "      <td>...</td>\n",
       "      <td>6.770754</td>\n",
       "      <td>6.606267</td>\n",
       "      <td>6.758579</td>\n",
       "      <td>6.685279</td>\n",
       "      <td>6.702444</td>\n",
       "      <td>6.548548</td>\n",
       "      <td>6.623991</td>\n",
       "      <td>6.503292</td>\n",
       "      <td>6.727856</td>\n",
       "      <td>6.595299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>9.911613</td>\n",
       "      <td>10.017467</td>\n",
       "      <td>9.871995</td>\n",
       "      <td>9.707972</td>\n",
       "      <td>10.086790</td>\n",
       "      <td>10.151445</td>\n",
       "      <td>10.318409</td>\n",
       "      <td>10.080975</td>\n",
       "      <td>10.183896</td>\n",
       "      <td>10.038278</td>\n",
       "      <td>...</td>\n",
       "      <td>9.724662</td>\n",
       "      <td>10.242638</td>\n",
       "      <td>9.792358</td>\n",
       "      <td>9.719300</td>\n",
       "      <td>10.147318</td>\n",
       "      <td>10.715561</td>\n",
       "      <td>10.077286</td>\n",
       "      <td>9.726975</td>\n",
       "      <td>10.139342</td>\n",
       "      <td>10.125031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>10.858828</td>\n",
       "      <td>8.512090</td>\n",
       "      <td>8.711756</td>\n",
       "      <td>9.209957</td>\n",
       "      <td>8.739951</td>\n",
       "      <td>8.567852</td>\n",
       "      <td>9.859342</td>\n",
       "      <td>9.021956</td>\n",
       "      <td>9.131341</td>\n",
       "      <td>8.492129</td>\n",
       "      <td>...</td>\n",
       "      <td>8.395954</td>\n",
       "      <td>8.109818</td>\n",
       "      <td>8.257713</td>\n",
       "      <td>9.571841</td>\n",
       "      <td>8.877332</td>\n",
       "      <td>8.493656</td>\n",
       "      <td>8.613942</td>\n",
       "      <td>8.916826</td>\n",
       "      <td>8.144795</td>\n",
       "      <td>8.406007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>8.413154</td>\n",
       "      <td>8.307626</td>\n",
       "      <td>8.234364</td>\n",
       "      <td>8.298999</td>\n",
       "      <td>8.282404</td>\n",
       "      <td>8.633262</td>\n",
       "      <td>8.541087</td>\n",
       "      <td>8.412907</td>\n",
       "      <td>8.163264</td>\n",
       "      <td>8.233234</td>\n",
       "      <td>...</td>\n",
       "      <td>8.620517</td>\n",
       "      <td>8.503926</td>\n",
       "      <td>8.273694</td>\n",
       "      <td>8.654892</td>\n",
       "      <td>8.389683</td>\n",
       "      <td>8.012815</td>\n",
       "      <td>8.396138</td>\n",
       "      <td>8.400772</td>\n",
       "      <td>8.212944</td>\n",
       "      <td>8.314829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>8.525627</td>\n",
       "      <td>8.723545</td>\n",
       "      <td>8.754528</td>\n",
       "      <td>8.591482</td>\n",
       "      <td>8.868986</td>\n",
       "      <td>9.154676</td>\n",
       "      <td>8.968219</td>\n",
       "      <td>8.918800</td>\n",
       "      <td>8.966573</td>\n",
       "      <td>8.646385</td>\n",
       "      <td>...</td>\n",
       "      <td>8.854108</td>\n",
       "      <td>8.960751</td>\n",
       "      <td>9.038261</td>\n",
       "      <td>9.150526</td>\n",
       "      <td>8.888503</td>\n",
       "      <td>8.894667</td>\n",
       "      <td>8.722772</td>\n",
       "      <td>8.783000</td>\n",
       "      <td>8.705091</td>\n",
       "      <td>8.872893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB1C</th>\n",
       "      <td>6.720996</td>\n",
       "      <td>6.614096</td>\n",
       "      <td>6.592334</td>\n",
       "      <td>6.602132</td>\n",
       "      <td>6.775020</td>\n",
       "      <td>6.598250</td>\n",
       "      <td>6.731486</td>\n",
       "      <td>6.685775</td>\n",
       "      <td>6.855869</td>\n",
       "      <td>6.709597</td>\n",
       "      <td>...</td>\n",
       "      <td>6.507397</td>\n",
       "      <td>6.568342</td>\n",
       "      <td>6.536055</td>\n",
       "      <td>6.523492</td>\n",
       "      <td>6.719801</td>\n",
       "      <td>6.596157</td>\n",
       "      <td>6.563001</td>\n",
       "      <td>6.661075</td>\n",
       "      <td>6.639940</td>\n",
       "      <td>6.565698</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17851 rows Ã— 242 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM2690230  GSM2690231  GSM2690232  GSM2690233  GSM2690234  \\\n",
       "A1BG       6.848050    6.688162    6.796304    6.726489    6.786394   \n",
       "A1CF       6.583041    6.650677    6.686538    6.677932    6.661399   \n",
       "A2M        6.650136    6.521780    6.566705    6.618576    6.580874   \n",
       "A2ML1      6.611510    6.646145    6.633586    6.460136    6.465241   \n",
       "A3GALT2    6.527584    6.622752    6.591259    6.624255    6.512508   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B     9.911613   10.017467    9.871995    9.707972   10.086790   \n",
       "ZYX       10.858828    8.512090    8.711756    9.209957    8.739951   \n",
       "ZZEF1      8.413154    8.307626    8.234364    8.298999    8.282404   \n",
       "ZZZ3       8.525627    8.723545    8.754528    8.591482    8.868986   \n",
       "RAB1C      6.720996    6.614096    6.592334    6.602132    6.775020   \n",
       "\n",
       "         GSM2690235  GSM2690236  GSM2690237  GSM2690238  GSM2690239  ...  \\\n",
       "A1BG       6.854404    6.674512    6.688926    6.807333    6.666278  ...   \n",
       "A1CF       6.609482    6.628126    6.613500    6.613940    6.596914  ...   \n",
       "A2M        6.569995    6.588907    6.559197    6.623999    6.567076  ...   \n",
       "A2ML1      6.422501    6.467049    6.607412    6.529374    6.381183  ...   \n",
       "A3GALT2    6.553509    6.691260    6.716285    6.605287    6.574475  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZYG11B    10.151445   10.318409   10.080975   10.183896   10.038278  ...   \n",
       "ZYX        8.567852    9.859342    9.021956    9.131341    8.492129  ...   \n",
       "ZZEF1      8.633262    8.541087    8.412907    8.163264    8.233234  ...   \n",
       "ZZZ3       9.154676    8.968219    8.918800    8.966573    8.646385  ...   \n",
       "RAB1C      6.598250    6.731486    6.685775    6.855869    6.709597  ...   \n",
       "\n",
       "         GSM2690462  GSM2690463  GSM2690464  GSM2690465  GSM2690466  \\\n",
       "A1BG       6.783802    6.657357    6.717128    6.735756    6.748284   \n",
       "A1CF       6.574616    6.561873    6.650160    6.606254    6.663059   \n",
       "A2M        6.633357    6.453508    6.538286    6.437383    6.601649   \n",
       "A2ML1      6.454338    6.698004    6.722807    6.648967    6.501302   \n",
       "A3GALT2    6.770754    6.606267    6.758579    6.685279    6.702444   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B     9.724662   10.242638    9.792358    9.719300   10.147318   \n",
       "ZYX        8.395954    8.109818    8.257713    9.571841    8.877332   \n",
       "ZZEF1      8.620517    8.503926    8.273694    8.654892    8.389683   \n",
       "ZZZ3       8.854108    8.960751    9.038261    9.150526    8.888503   \n",
       "RAB1C      6.507397    6.568342    6.536055    6.523492    6.719801   \n",
       "\n",
       "         GSM2690467  GSM2690468  GSM2690469  GSM2690470  GSM2690471  \n",
       "A1BG       6.761963    6.747374    6.804050    6.778584    6.751946  \n",
       "A1CF       6.624456    6.657874    6.672318    6.741131    6.641459  \n",
       "A2M        6.726172    6.425379    6.620081    6.528318    6.530385  \n",
       "A2ML1      6.586460    6.566908    6.641582    6.606379    6.488792  \n",
       "A3GALT2    6.548548    6.623991    6.503292    6.727856    6.595299  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZYG11B    10.715561   10.077286    9.726975   10.139342   10.125031  \n",
       "ZYX        8.493656    8.613942    8.916826    8.144795    8.406007  \n",
       "ZZEF1      8.012815    8.396138    8.400772    8.212944    8.314829  \n",
       "ZZZ3       8.894667    8.722772    8.783000    8.705091    8.872893  \n",
       "RAB1C      6.596157    6.563001    6.661075    6.639940    6.565698  \n",
       "\n",
       "[17851 rows x 242 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>RAB1C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2690230</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.848050</td>\n",
       "      <td>6.583041</td>\n",
       "      <td>6.650136</td>\n",
       "      <td>6.611510</td>\n",
       "      <td>6.527584</td>\n",
       "      <td>6.434655</td>\n",
       "      <td>7.058692</td>\n",
       "      <td>6.700324</td>\n",
       "      <td>...</td>\n",
       "      <td>6.880832</td>\n",
       "      <td>7.006752</td>\n",
       "      <td>7.284013</td>\n",
       "      <td>6.975131</td>\n",
       "      <td>6.629065</td>\n",
       "      <td>9.911613</td>\n",
       "      <td>10.858828</td>\n",
       "      <td>8.413154</td>\n",
       "      <td>8.525627</td>\n",
       "      <td>6.720996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690231</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.688162</td>\n",
       "      <td>6.650677</td>\n",
       "      <td>6.521780</td>\n",
       "      <td>6.646145</td>\n",
       "      <td>6.622752</td>\n",
       "      <td>6.618596</td>\n",
       "      <td>6.880522</td>\n",
       "      <td>6.620599</td>\n",
       "      <td>...</td>\n",
       "      <td>6.560931</td>\n",
       "      <td>7.133329</td>\n",
       "      <td>7.969672</td>\n",
       "      <td>7.378143</td>\n",
       "      <td>6.566949</td>\n",
       "      <td>10.017467</td>\n",
       "      <td>8.512090</td>\n",
       "      <td>8.307626</td>\n",
       "      <td>8.723545</td>\n",
       "      <td>6.614096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690232</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.796304</td>\n",
       "      <td>6.686538</td>\n",
       "      <td>6.566705</td>\n",
       "      <td>6.633586</td>\n",
       "      <td>6.591259</td>\n",
       "      <td>6.555226</td>\n",
       "      <td>6.825927</td>\n",
       "      <td>6.636943</td>\n",
       "      <td>...</td>\n",
       "      <td>6.619013</td>\n",
       "      <td>7.188181</td>\n",
       "      <td>7.688697</td>\n",
       "      <td>7.205052</td>\n",
       "      <td>6.687631</td>\n",
       "      <td>9.871995</td>\n",
       "      <td>8.711756</td>\n",
       "      <td>8.234364</td>\n",
       "      <td>8.754528</td>\n",
       "      <td>6.592334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690233</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.726489</td>\n",
       "      <td>6.677932</td>\n",
       "      <td>6.618576</td>\n",
       "      <td>6.460136</td>\n",
       "      <td>6.624255</td>\n",
       "      <td>6.605685</td>\n",
       "      <td>6.989004</td>\n",
       "      <td>6.624586</td>\n",
       "      <td>...</td>\n",
       "      <td>6.705561</td>\n",
       "      <td>7.072152</td>\n",
       "      <td>7.534276</td>\n",
       "      <td>7.056405</td>\n",
       "      <td>6.583996</td>\n",
       "      <td>9.707972</td>\n",
       "      <td>9.209957</td>\n",
       "      <td>8.298999</td>\n",
       "      <td>8.591482</td>\n",
       "      <td>6.602132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690234</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.786394</td>\n",
       "      <td>6.661399</td>\n",
       "      <td>6.580874</td>\n",
       "      <td>6.465241</td>\n",
       "      <td>6.512508</td>\n",
       "      <td>6.698219</td>\n",
       "      <td>6.666786</td>\n",
       "      <td>6.656601</td>\n",
       "      <td>...</td>\n",
       "      <td>6.697169</td>\n",
       "      <td>7.046910</td>\n",
       "      <td>7.638880</td>\n",
       "      <td>7.310057</td>\n",
       "      <td>6.686599</td>\n",
       "      <td>10.086790</td>\n",
       "      <td>8.739951</td>\n",
       "      <td>8.282404</td>\n",
       "      <td>8.868986</td>\n",
       "      <td>6.775020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690467</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.761963</td>\n",
       "      <td>6.624456</td>\n",
       "      <td>6.726172</td>\n",
       "      <td>6.586460</td>\n",
       "      <td>6.548548</td>\n",
       "      <td>6.567630</td>\n",
       "      <td>6.679969</td>\n",
       "      <td>6.674690</td>\n",
       "      <td>...</td>\n",
       "      <td>6.979054</td>\n",
       "      <td>7.193730</td>\n",
       "      <td>8.066666</td>\n",
       "      <td>7.221787</td>\n",
       "      <td>6.595565</td>\n",
       "      <td>10.715561</td>\n",
       "      <td>8.493656</td>\n",
       "      <td>8.012815</td>\n",
       "      <td>8.894667</td>\n",
       "      <td>6.596157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690468</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.747374</td>\n",
       "      <td>6.657874</td>\n",
       "      <td>6.425379</td>\n",
       "      <td>6.566908</td>\n",
       "      <td>6.623991</td>\n",
       "      <td>6.576486</td>\n",
       "      <td>6.668170</td>\n",
       "      <td>6.618987</td>\n",
       "      <td>...</td>\n",
       "      <td>6.836740</td>\n",
       "      <td>7.007840</td>\n",
       "      <td>7.708474</td>\n",
       "      <td>7.048696</td>\n",
       "      <td>6.667023</td>\n",
       "      <td>10.077286</td>\n",
       "      <td>8.613942</td>\n",
       "      <td>8.396138</td>\n",
       "      <td>8.722772</td>\n",
       "      <td>6.563001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690469</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.804050</td>\n",
       "      <td>6.672318</td>\n",
       "      <td>6.620081</td>\n",
       "      <td>6.641582</td>\n",
       "      <td>6.503292</td>\n",
       "      <td>6.665137</td>\n",
       "      <td>6.746265</td>\n",
       "      <td>6.649298</td>\n",
       "      <td>...</td>\n",
       "      <td>6.718197</td>\n",
       "      <td>7.087727</td>\n",
       "      <td>7.800900</td>\n",
       "      <td>7.160791</td>\n",
       "      <td>6.633927</td>\n",
       "      <td>9.726975</td>\n",
       "      <td>8.916826</td>\n",
       "      <td>8.400772</td>\n",
       "      <td>8.783000</td>\n",
       "      <td>6.661075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690470</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.778584</td>\n",
       "      <td>6.741131</td>\n",
       "      <td>6.528318</td>\n",
       "      <td>6.606379</td>\n",
       "      <td>6.727856</td>\n",
       "      <td>6.466134</td>\n",
       "      <td>6.547827</td>\n",
       "      <td>6.558552</td>\n",
       "      <td>...</td>\n",
       "      <td>6.751371</td>\n",
       "      <td>7.033584</td>\n",
       "      <td>7.752131</td>\n",
       "      <td>6.960197</td>\n",
       "      <td>6.550888</td>\n",
       "      <td>10.139342</td>\n",
       "      <td>8.144795</td>\n",
       "      <td>8.212944</td>\n",
       "      <td>8.705091</td>\n",
       "      <td>6.639940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2690471</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.751946</td>\n",
       "      <td>6.641459</td>\n",
       "      <td>6.530385</td>\n",
       "      <td>6.488792</td>\n",
       "      <td>6.595299</td>\n",
       "      <td>6.678781</td>\n",
       "      <td>6.649477</td>\n",
       "      <td>6.629344</td>\n",
       "      <td>...</td>\n",
       "      <td>6.662896</td>\n",
       "      <td>7.147226</td>\n",
       "      <td>7.946082</td>\n",
       "      <td>7.175329</td>\n",
       "      <td>6.694830</td>\n",
       "      <td>10.125031</td>\n",
       "      <td>8.406007</td>\n",
       "      <td>8.314829</td>\n",
       "      <td>8.872893</td>\n",
       "      <td>6.565698</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>242 rows Ã— 17853 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Ankylosing Spondylitis  Gender      A1BG      A1CF       A2M  \\\n",
       "GSM2690230                     0.0     0.0  6.848050  6.583041  6.650136   \n",
       "GSM2690231                     0.0     1.0  6.688162  6.650677  6.521780   \n",
       "GSM2690232                     0.0     1.0  6.796304  6.686538  6.566705   \n",
       "GSM2690233                     0.0     1.0  6.726489  6.677932  6.618576   \n",
       "GSM2690234                     0.0     0.0  6.786394  6.661399  6.580874   \n",
       "...                            ...     ...       ...       ...       ...   \n",
       "GSM2690467                     0.0     0.0  6.761963  6.624456  6.726172   \n",
       "GSM2690468                     0.0     1.0  6.747374  6.657874  6.425379   \n",
       "GSM2690469                     0.0     1.0  6.804050  6.672318  6.620081   \n",
       "GSM2690470                     0.0     1.0  6.778584  6.741131  6.528318   \n",
       "GSM2690471                     0.0     1.0  6.751946  6.641459  6.530385   \n",
       "\n",
       "               A2ML1   A3GALT2    A4GALT     A4GNT      AAA1  ...     ZWINT  \\\n",
       "GSM2690230  6.611510  6.527584  6.434655  7.058692  6.700324  ...  6.880832   \n",
       "GSM2690231  6.646145  6.622752  6.618596  6.880522  6.620599  ...  6.560931   \n",
       "GSM2690232  6.633586  6.591259  6.555226  6.825927  6.636943  ...  6.619013   \n",
       "GSM2690233  6.460136  6.624255  6.605685  6.989004  6.624586  ...  6.705561   \n",
       "GSM2690234  6.465241  6.512508  6.698219  6.666786  6.656601  ...  6.697169   \n",
       "...              ...       ...       ...       ...       ...  ...       ...   \n",
       "GSM2690467  6.586460  6.548548  6.567630  6.679969  6.674690  ...  6.979054   \n",
       "GSM2690468  6.566908  6.623991  6.576486  6.668170  6.618987  ...  6.836740   \n",
       "GSM2690469  6.641582  6.503292  6.665137  6.746265  6.649298  ...  6.718197   \n",
       "GSM2690470  6.606379  6.727856  6.466134  6.547827  6.558552  ...  6.751371   \n",
       "GSM2690471  6.488792  6.595299  6.678781  6.649477  6.629344  ...  6.662896   \n",
       "\n",
       "                ZXDA      ZXDB      ZXDC    ZYG11A     ZYG11B        ZYX  \\\n",
       "GSM2690230  7.006752  7.284013  6.975131  6.629065   9.911613  10.858828   \n",
       "GSM2690231  7.133329  7.969672  7.378143  6.566949  10.017467   8.512090   \n",
       "GSM2690232  7.188181  7.688697  7.205052  6.687631   9.871995   8.711756   \n",
       "GSM2690233  7.072152  7.534276  7.056405  6.583996   9.707972   9.209957   \n",
       "GSM2690234  7.046910  7.638880  7.310057  6.686599  10.086790   8.739951   \n",
       "...              ...       ...       ...       ...        ...        ...   \n",
       "GSM2690467  7.193730  8.066666  7.221787  6.595565  10.715561   8.493656   \n",
       "GSM2690468  7.007840  7.708474  7.048696  6.667023  10.077286   8.613942   \n",
       "GSM2690469  7.087727  7.800900  7.160791  6.633927   9.726975   8.916826   \n",
       "GSM2690470  7.033584  7.752131  6.960197  6.550888  10.139342   8.144795   \n",
       "GSM2690471  7.147226  7.946082  7.175329  6.694830  10.125031   8.406007   \n",
       "\n",
       "               ZZEF1      ZZZ3     RAB1C  \n",
       "GSM2690230  8.413154  8.525627  6.720996  \n",
       "GSM2690231  8.307626  8.723545  6.614096  \n",
       "GSM2690232  8.234364  8.754528  6.592334  \n",
       "GSM2690233  8.298999  8.591482  6.602132  \n",
       "GSM2690234  8.282404  8.868986  6.775020  \n",
       "...              ...       ...       ...  \n",
       "GSM2690467  8.012815  8.894667  6.596157  \n",
       "GSM2690468  8.396138  8.722772  6.563001  \n",
       "GSM2690469  8.400772  8.783000  6.661075  \n",
       "GSM2690470  8.212944  8.705091  6.639940  \n",
       "GSM2690471  8.314829  8.872893  6.565698  \n",
       "\n",
       "[242 rows x 17853 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 242 samples.\n",
      "For the feature 'Ankylosing Spondylitis', the least common label is '1.0' with 2 occurrences. This represents 0.83% of the dataset.\n",
      "The distribution of the feature 'Ankylosing Spondylitis' in this dataset is severely biased.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 69 occurrences. This represents 28.51% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Expression profiling in whole blood in ankylosing spondylitis patients and controls\"\n",
      "!Series_summary\t\"Introduction: A number of genetic-association studies have identified genes contributing to AS susceptibility but such approaches provide little information as to the gene activity changes occurring during the disease process. Transcriptional profiling generates a â€œsnapshotâ€ of the sampled cells activity and thus can provide insights into the molecular processes driving the disease process. We undertook a whole-genome microarray approach to identify candidate genes associated with AS and validated these gene-expression changes in a larger sample cohort.  Methods: 18 active AS patients, classified according to the New York criteria. and 18 gender-and age-matched controls were profiled using Illumina HT-12 Whole-Genome Expression BeadChips which carry cDNAs for 48000 genes and transcripts. Class comparison analysis identified a number of differentially expressed candidate genes. These candidate genes were then validated in a larger cohort using qPCR-based TaqMan Low Density Arrays (TLDAs).  Results: 239 probes corresponding to 221 genes were identified as being significantly different between patients and controls with a p-value <0.0005 (80% confidence level of false discovery rate). Forty seven genes were then selected for validation studies, using the TLDAs. Thirteen of these genes were validated in the second patient cohort with 12 down-regulated 1.3-2-fold and only 1 upregulated (1.6-fold). Among a number of identified genes with well-documented inflammatory roles we also validated genes that might be of great interest to the understanding of AS progression such as SPOCK2 (osteonectin) and EP300 which modulate cartilage and bone metabolism.  Conclusion: We have validated a gene expression signature for AS from whole blood and identified strong candidate genes that may play roles in both the inflammatory and joint destruction aspects of the disease.\"\n",
      "!Series_overall_design\t\"RNA was extracted from whole blood using PAXGene tubes. 16 AS patients with active disease and 16 gender- and age-matched controls were analysed.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM616668</th>\n",
       "      <th>GSM616669</th>\n",
       "      <th>GSM616670</th>\n",
       "      <th>GSM616671</th>\n",
       "      <th>GSM616672</th>\n",
       "      <th>GSM616673</th>\n",
       "      <th>GSM616674</th>\n",
       "      <th>GSM616675</th>\n",
       "      <th>GSM616676</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM616690</th>\n",
       "      <th>GSM616691</th>\n",
       "      <th>GSM616692</th>\n",
       "      <th>GSM616693</th>\n",
       "      <th>GSM616694</th>\n",
       "      <th>GSM616695</th>\n",
       "      <th>GSM616696</th>\n",
       "      <th>GSM616697</th>\n",
       "      <th>GSM616698</th>\n",
       "      <th>GSM616699</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "      <td>tissue: Whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "      <td>cell type: PBMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>disease status: Ankylosing spondylitis patient</td>\n",
       "      <td>...</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "      <td>disease status: Normal control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "\n",
       "                                        GSM616668  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616669  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616670  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616671  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616672  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616673  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616674  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616675  \\\n",
       "0                             tissue: Whole blood   \n",
       "1                                 cell type: PBMC   \n",
       "2  disease status: Ankylosing spondylitis patient   \n",
       "\n",
       "                                        GSM616676  ...  \\\n",
       "0                             tissue: Whole blood  ...   \n",
       "1                                 cell type: PBMC  ...   \n",
       "2  disease status: Ankylosing spondylitis patient  ...   \n",
       "\n",
       "                        GSM616690                       GSM616691  \\\n",
       "0             tissue: Whole blood             tissue: Whole blood   \n",
       "1                 cell type: PBMC                 cell type: PBMC   \n",
       "2  disease status: Normal control  disease status: Normal control   \n",
       "\n",
       "                        GSM616692                       GSM616693  \\\n",
       "0             tissue: Whole blood             tissue: Whole blood   \n",
       "1                 cell type: PBMC                 cell type: PBMC   \n",
       "2  disease status: Normal control  disease status: Normal control   \n",
       "\n",
       "                        GSM616694                       GSM616695  \\\n",
       "0             tissue: Whole blood             tissue: Whole blood   \n",
       "1                 cell type: PBMC                 cell type: PBMC   \n",
       "2  disease status: Normal control  disease status: Normal control   \n",
       "\n",
       "                        GSM616696                       GSM616697  \\\n",
       "0             tissue: Whole blood             tissue: Whole blood   \n",
       "1                 cell type: PBMC                 cell type: PBMC   \n",
       "2  disease status: Normal control  disease status: Normal control   \n",
       "\n",
       "                        GSM616698                       GSM616699  \n",
       "0             tissue: Whole blood             tissue: Whole blood  \n",
       "1                 cell type: PBMC                 cell type: PBMC  \n",
       "2  disease status: Normal control  disease status: Normal control  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE25101\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'disease status: Ankylosing spondylitis patient',\n",
       "       'disease status: Normal control'], dtype=object)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease status: Ankylosing spondylitis patient':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM616668</th>\n",
       "      <th>GSM616669</th>\n",
       "      <th>GSM616670</th>\n",
       "      <th>GSM616671</th>\n",
       "      <th>GSM616672</th>\n",
       "      <th>GSM616673</th>\n",
       "      <th>GSM616674</th>\n",
       "      <th>GSM616675</th>\n",
       "      <th>GSM616676</th>\n",
       "      <th>GSM616677</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM616690</th>\n",
       "      <th>GSM616691</th>\n",
       "      <th>GSM616692</th>\n",
       "      <th>GSM616693</th>\n",
       "      <th>GSM616694</th>\n",
       "      <th>GSM616695</th>\n",
       "      <th>GSM616696</th>\n",
       "      <th>GSM616697</th>\n",
       "      <th>GSM616698</th>\n",
       "      <th>GSM616699</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GSM616668  GSM616669  GSM616670  GSM616671  GSM616672  \\\n",
       "Ankylosing Spondylitis          1          1          1          1          1   \n",
       "\n",
       "                        GSM616673  GSM616674  GSM616675  GSM616676  GSM616677  \\\n",
       "Ankylosing Spondylitis          1          1          1          1          1   \n",
       "\n",
       "                        ...  GSM616690  GSM616691  GSM616692  GSM616693  \\\n",
       "Ankylosing Spondylitis  ...          0          0          0          0   \n",
       "\n",
       "                        GSM616694  GSM616695  GSM616696  GSM616697  GSM616698  \\\n",
       "Ankylosing Spondylitis          0          0          0          0          0   \n",
       "\n",
       "                        GSM616699  \n",
       "Ankylosing Spondylitis          0  \n",
       "\n",
       "[1 rows x 32 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM616668</th>\n",
       "      <th>GSM616669</th>\n",
       "      <th>GSM616670</th>\n",
       "      <th>GSM616671</th>\n",
       "      <th>GSM616672</th>\n",
       "      <th>GSM616673</th>\n",
       "      <th>GSM616674</th>\n",
       "      <th>GSM616675</th>\n",
       "      <th>GSM616676</th>\n",
       "      <th>GSM616677</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM616690</th>\n",
       "      <th>GSM616691</th>\n",
       "      <th>GSM616692</th>\n",
       "      <th>GSM616693</th>\n",
       "      <th>GSM616694</th>\n",
       "      <th>GSM616695</th>\n",
       "      <th>GSM616696</th>\n",
       "      <th>GSM616697</th>\n",
       "      <th>GSM616698</th>\n",
       "      <th>GSM616699</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>14.721687</td>\n",
       "      <td>14.910628</td>\n",
       "      <td>14.945053</td>\n",
       "      <td>14.868905</td>\n",
       "      <td>14.910628</td>\n",
       "      <td>14.994942</td>\n",
       "      <td>14.579192</td>\n",
       "      <td>14.889235</td>\n",
       "      <td>14.868905</td>\n",
       "      <td>14.789278</td>\n",
       "      <td>...</td>\n",
       "      <td>14.889235</td>\n",
       "      <td>14.889235</td>\n",
       "      <td>14.945053</td>\n",
       "      <td>14.683020</td>\n",
       "      <td>14.817801</td>\n",
       "      <td>14.910628</td>\n",
       "      <td>14.889235</td>\n",
       "      <td>14.776362</td>\n",
       "      <td>14.737311</td>\n",
       "      <td>14.868905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1343295</th>\n",
       "      <td>11.818846</td>\n",
       "      <td>12.014962</td>\n",
       "      <td>12.020319</td>\n",
       "      <td>11.923485</td>\n",
       "      <td>11.674407</td>\n",
       "      <td>11.829631</td>\n",
       "      <td>11.492026</td>\n",
       "      <td>12.027219</td>\n",
       "      <td>11.738941</td>\n",
       "      <td>12.599862</td>\n",
       "      <td>...</td>\n",
       "      <td>12.233341</td>\n",
       "      <td>11.936598</td>\n",
       "      <td>12.282542</td>\n",
       "      <td>11.938787</td>\n",
       "      <td>11.631719</td>\n",
       "      <td>11.804583</td>\n",
       "      <td>11.701640</td>\n",
       "      <td>11.929974</td>\n",
       "      <td>11.738941</td>\n",
       "      <td>11.578072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651209</th>\n",
       "      <td>6.869540</td>\n",
       "      <td>7.095127</td>\n",
       "      <td>7.031877</td>\n",
       "      <td>7.133220</td>\n",
       "      <td>6.887904</td>\n",
       "      <td>6.918006</td>\n",
       "      <td>6.894105</td>\n",
       "      <td>6.886695</td>\n",
       "      <td>7.016599</td>\n",
       "      <td>6.924489</td>\n",
       "      <td>...</td>\n",
       "      <td>6.694988</td>\n",
       "      <td>6.839130</td>\n",
       "      <td>7.323969</td>\n",
       "      <td>6.973208</td>\n",
       "      <td>6.868359</td>\n",
       "      <td>7.265285</td>\n",
       "      <td>7.415702</td>\n",
       "      <td>7.236032</td>\n",
       "      <td>7.022915</td>\n",
       "      <td>7.205181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651228</th>\n",
       "      <td>11.300704</td>\n",
       "      <td>11.486897</td>\n",
       "      <td>11.337098</td>\n",
       "      <td>12.668199</td>\n",
       "      <td>11.107247</td>\n",
       "      <td>11.330551</td>\n",
       "      <td>12.863078</td>\n",
       "      <td>12.654738</td>\n",
       "      <td>12.784749</td>\n",
       "      <td>12.150922</td>\n",
       "      <td>...</td>\n",
       "      <td>10.968782</td>\n",
       "      <td>11.407495</td>\n",
       "      <td>13.174214</td>\n",
       "      <td>12.512700</td>\n",
       "      <td>13.083279</td>\n",
       "      <td>12.814878</td>\n",
       "      <td>12.476037</td>\n",
       "      <td>12.498150</td>\n",
       "      <td>12.313310</td>\n",
       "      <td>11.912799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651229</th>\n",
       "      <td>7.687854</td>\n",
       "      <td>8.152096</td>\n",
       "      <td>7.854758</td>\n",
       "      <td>7.793770</td>\n",
       "      <td>7.773122</td>\n",
       "      <td>7.649979</td>\n",
       "      <td>7.808541</td>\n",
       "      <td>7.788769</td>\n",
       "      <td>8.079810</td>\n",
       "      <td>8.140961</td>\n",
       "      <td>...</td>\n",
       "      <td>8.059401</td>\n",
       "      <td>7.950677</td>\n",
       "      <td>8.487722</td>\n",
       "      <td>8.039596</td>\n",
       "      <td>7.828835</td>\n",
       "      <td>8.255879</td>\n",
       "      <td>8.148248</td>\n",
       "      <td>8.321234</td>\n",
       "      <td>7.813087</td>\n",
       "      <td>7.871922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415898</th>\n",
       "      <td>7.976265</td>\n",
       "      <td>7.467628</td>\n",
       "      <td>7.779078</td>\n",
       "      <td>7.799240</td>\n",
       "      <td>7.642209</td>\n",
       "      <td>7.737723</td>\n",
       "      <td>8.040922</td>\n",
       "      <td>8.059585</td>\n",
       "      <td>7.551395</td>\n",
       "      <td>7.646354</td>\n",
       "      <td>...</td>\n",
       "      <td>7.673213</td>\n",
       "      <td>7.906765</td>\n",
       "      <td>7.839494</td>\n",
       "      <td>7.832799</td>\n",
       "      <td>8.817948</td>\n",
       "      <td>7.376251</td>\n",
       "      <td>8.010885</td>\n",
       "      <td>7.484848</td>\n",
       "      <td>8.145814</td>\n",
       "      <td>8.037064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415911</th>\n",
       "      <td>7.753634</td>\n",
       "      <td>7.728451</td>\n",
       "      <td>7.741349</td>\n",
       "      <td>7.352076</td>\n",
       "      <td>7.736549</td>\n",
       "      <td>7.857763</td>\n",
       "      <td>7.634732</td>\n",
       "      <td>7.350687</td>\n",
       "      <td>7.762689</td>\n",
       "      <td>7.677580</td>\n",
       "      <td>...</td>\n",
       "      <td>7.860533</td>\n",
       "      <td>7.655432</td>\n",
       "      <td>7.467160</td>\n",
       "      <td>7.627300</td>\n",
       "      <td>7.769050</td>\n",
       "      <td>7.573400</td>\n",
       "      <td>7.615588</td>\n",
       "      <td>7.719455</td>\n",
       "      <td>7.598297</td>\n",
       "      <td>7.516465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415926</th>\n",
       "      <td>7.812243</td>\n",
       "      <td>7.828331</td>\n",
       "      <td>8.236525</td>\n",
       "      <td>8.052980</td>\n",
       "      <td>7.933305</td>\n",
       "      <td>8.180785</td>\n",
       "      <td>8.112408</td>\n",
       "      <td>7.564648</td>\n",
       "      <td>8.198021</td>\n",
       "      <td>7.351243</td>\n",
       "      <td>...</td>\n",
       "      <td>7.972885</td>\n",
       "      <td>8.229952</td>\n",
       "      <td>7.826864</td>\n",
       "      <td>7.976265</td>\n",
       "      <td>7.928434</td>\n",
       "      <td>8.080596</td>\n",
       "      <td>8.407442</td>\n",
       "      <td>7.698175</td>\n",
       "      <td>8.157596</td>\n",
       "      <td>7.569014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415949</th>\n",
       "      <td>7.937806</td>\n",
       "      <td>7.797597</td>\n",
       "      <td>7.981566</td>\n",
       "      <td>8.007248</td>\n",
       "      <td>7.969345</td>\n",
       "      <td>8.101712</td>\n",
       "      <td>7.859426</td>\n",
       "      <td>7.683960</td>\n",
       "      <td>8.171502</td>\n",
       "      <td>7.814201</td>\n",
       "      <td>...</td>\n",
       "      <td>7.941449</td>\n",
       "      <td>7.864130</td>\n",
       "      <td>7.956597</td>\n",
       "      <td>7.738982</td>\n",
       "      <td>7.996923</td>\n",
       "      <td>8.098316</td>\n",
       "      <td>7.850872</td>\n",
       "      <td>7.967910</td>\n",
       "      <td>8.114295</td>\n",
       "      <td>7.848682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415979</th>\n",
       "      <td>10.976353</td>\n",
       "      <td>12.084026</td>\n",
       "      <td>11.529173</td>\n",
       "      <td>12.056335</td>\n",
       "      <td>11.797443</td>\n",
       "      <td>10.891747</td>\n",
       "      <td>12.169497</td>\n",
       "      <td>11.470129</td>\n",
       "      <td>12.148157</td>\n",
       "      <td>11.478385</td>\n",
       "      <td>...</td>\n",
       "      <td>9.953578</td>\n",
       "      <td>11.804583</td>\n",
       "      <td>12.252931</td>\n",
       "      <td>11.967590</td>\n",
       "      <td>11.740767</td>\n",
       "      <td>12.310597</td>\n",
       "      <td>12.077604</td>\n",
       "      <td>12.200691</td>\n",
       "      <td>11.779882</td>\n",
       "      <td>12.259073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18168 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM616668  GSM616669  GSM616670  GSM616671  GSM616672  \\\n",
       "ID                                                                    \n",
       "ILMN_1343291  14.721687  14.910628  14.945053  14.868905  14.910628   \n",
       "ILMN_1343295  11.818846  12.014962  12.020319  11.923485  11.674407   \n",
       "ILMN_1651209   6.869540   7.095127   7.031877   7.133220   6.887904   \n",
       "ILMN_1651228  11.300704  11.486897  11.337098  12.668199  11.107247   \n",
       "ILMN_1651229   7.687854   8.152096   7.854758   7.793770   7.773122   \n",
       "...                 ...        ...        ...        ...        ...   \n",
       "ILMN_2415898   7.976265   7.467628   7.779078   7.799240   7.642209   \n",
       "ILMN_2415911   7.753634   7.728451   7.741349   7.352076   7.736549   \n",
       "ILMN_2415926   7.812243   7.828331   8.236525   8.052980   7.933305   \n",
       "ILMN_2415949   7.937806   7.797597   7.981566   8.007248   7.969345   \n",
       "ILMN_2415979  10.976353  12.084026  11.529173  12.056335  11.797443   \n",
       "\n",
       "              GSM616673  GSM616674  GSM616675  GSM616676  GSM616677  ...  \\\n",
       "ID                                                                   ...   \n",
       "ILMN_1343291  14.994942  14.579192  14.889235  14.868905  14.789278  ...   \n",
       "ILMN_1343295  11.829631  11.492026  12.027219  11.738941  12.599862  ...   \n",
       "ILMN_1651209   6.918006   6.894105   6.886695   7.016599   6.924489  ...   \n",
       "ILMN_1651228  11.330551  12.863078  12.654738  12.784749  12.150922  ...   \n",
       "ILMN_1651229   7.649979   7.808541   7.788769   8.079810   8.140961  ...   \n",
       "...                 ...        ...        ...        ...        ...  ...   \n",
       "ILMN_2415898   7.737723   8.040922   8.059585   7.551395   7.646354  ...   \n",
       "ILMN_2415911   7.857763   7.634732   7.350687   7.762689   7.677580  ...   \n",
       "ILMN_2415926   8.180785   8.112408   7.564648   8.198021   7.351243  ...   \n",
       "ILMN_2415949   8.101712   7.859426   7.683960   8.171502   7.814201  ...   \n",
       "ILMN_2415979  10.891747  12.169497  11.470129  12.148157  11.478385  ...   \n",
       "\n",
       "              GSM616690  GSM616691  GSM616692  GSM616693  GSM616694  \\\n",
       "ID                                                                    \n",
       "ILMN_1343291  14.889235  14.889235  14.945053  14.683020  14.817801   \n",
       "ILMN_1343295  12.233341  11.936598  12.282542  11.938787  11.631719   \n",
       "ILMN_1651209   6.694988   6.839130   7.323969   6.973208   6.868359   \n",
       "ILMN_1651228  10.968782  11.407495  13.174214  12.512700  13.083279   \n",
       "ILMN_1651229   8.059401   7.950677   8.487722   8.039596   7.828835   \n",
       "...                 ...        ...        ...        ...        ...   \n",
       "ILMN_2415898   7.673213   7.906765   7.839494   7.832799   8.817948   \n",
       "ILMN_2415911   7.860533   7.655432   7.467160   7.627300   7.769050   \n",
       "ILMN_2415926   7.972885   8.229952   7.826864   7.976265   7.928434   \n",
       "ILMN_2415949   7.941449   7.864130   7.956597   7.738982   7.996923   \n",
       "ILMN_2415979   9.953578  11.804583  12.252931  11.967590  11.740767   \n",
       "\n",
       "              GSM616695  GSM616696  GSM616697  GSM616698  GSM616699  \n",
       "ID                                                                   \n",
       "ILMN_1343291  14.910628  14.889235  14.776362  14.737311  14.868905  \n",
       "ILMN_1343295  11.804583  11.701640  11.929974  11.738941  11.578072  \n",
       "ILMN_1651209   7.265285   7.415702   7.236032   7.022915   7.205181  \n",
       "ILMN_1651228  12.814878  12.476037  12.498150  12.313310  11.912799  \n",
       "ILMN_1651229   8.255879   8.148248   8.321234   7.813087   7.871922  \n",
       "...                 ...        ...        ...        ...        ...  \n",
       "ILMN_2415898   7.376251   8.010885   7.484848   8.145814   8.037064  \n",
       "ILMN_2415911   7.573400   7.615588   7.719455   7.598297   7.516465  \n",
       "ILMN_2415926   8.080596   8.407442   7.698175   8.157596   7.569014  \n",
       "ILMN_2415949   8.098316   7.850872   7.967910   8.114295   7.848682  \n",
       "ILMN_2415979  12.310597  12.077604  12.200691  11.779882  12.259073  \n",
       "\n",
       "[18168 rows x 32 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_1725881', 'ILMN_1910180', 'ILMN_1804174', 'ILMN_1796063', 'ILMN_1811966'], 'nuID': ['rp13_p1x6D80lNLk3c', 'NEX0oqCV8.er4HVfU4', 'KyqQynMZxJcruyylEU', 'xXl7eXuF7sbPEp.KFI', '9ckqJrioiaej9_ajeQ'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['RefSeq', 'Unigene', 'RefSeq', 'RefSeq', 'RefSeq'], 'Search_Key': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'Transcript': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'ILMN_Gene': ['LOC23117', 'HS.575038', 'FCGR2B', 'TRIM44', 'LOC653895'], 'Source_Reference_ID': ['XM_933824.1', 'Hs.575038', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'RefSeq_ID': ['XM_933824.1', nan, 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Unigene_ID': [nan, 'Hs.575038', nan, nan, nan], 'Entrez_Gene_ID': [23117.0, nan, 2213.0, 54765.0, 653895.0], 'GI': [89040007.0, 10437021.0, 88952550.0, 29029528.0, 89033487.0], 'Accession': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Symbol': ['LOC23117', nan, 'FCGR2B', 'TRIM44', 'LOC653895'], 'Protein_Product': ['XP_938917.1', nan, 'XP_943944.1', 'NP_060053.2', 'XP_941472.1'], 'Array_Address_Id': [1710221.0, 5900364.0, 2480717.0, 1300239.0, 4480719.0], 'Probe_Type': ['I', 'S', 'I', 'S', 'S'], 'Probe_Start': [122.0, 1409.0, 1643.0, 2901.0, 25.0], 'SEQUENCE': ['GGCTCCTCTTTGGGCTCCTACTGGAATTTATCAGCCATCAGTGCATCTCT', 'ACACCTTCAGGAGGGAAGCCCTTATTTCTGGGTTGAACTCCCCTTCCATG', 'TAGGGGCAATAGGCTATACGCTACAGCCTAGGTGTGTAGTAGGCCACACC', 'CCTGCCTGTCTGCCTGTGACCTGTGTACGTATTACAGGCTTTAGGACCAG', 'CTAGCAGGGAGCGGTGAGGGAGAGCGGCTGGATTTCTTGCGGGATCTGCA'], 'Chromosome': ['16', nan, nan, '11', nan], 'Probe_Chr_Orientation': ['-', nan, nan, '+', nan], 'Probe_Coordinates': ['21766363-21766363:21769901-21769949', nan, nan, '35786070-35786119', nan], 'Cytoband': ['16p12.2a', nan, '1q23.3b', '11p13a', '10q11.23b'], 'Definition': ['PREDICTED: Homo sapiens KIAA0220-like protein, transcript variant 11 (LOC23117), mRNA.', 'Homo sapiens cDNA: FLJ21027 fis, clone CAE07110', 'PREDICTED: Homo sapiens Fc fragment of IgG, low affinity IIb, receptor (CD32) (FCGR2B), mRNA.', 'Homo sapiens tripartite motif-containing 44 (TRIM44), mRNA.', 'PREDICTED: Homo sapiens similar to protein geranylgeranyltransferase type I, beta subunit (LOC653895), mRNA.'], 'Ontology_Component': [nan, nan, nan, 'intracellular [goid 5622] [evidence IEA]', nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, 'zinc ion binding [goid 8270] [evidence IEA]; metal ion binding [goid 46872] [evidence IEA]', nan], 'Synonyms': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'Obsolete_Probe_Id': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'GB_ACC': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651209', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651232', 'ILMN_1651237', 'ILMN_1651254', 'ILMN_1651262', 'ILMN_1651268', 'ILMN_1651278', 'ILMN_1651282', 'ILMN_1651286', 'ILMN_1651296', 'ILMN_1651315', 'ILMN_1651316', 'ILMN_1651328', 'ILMN_1651336', 'ILMN_1651341', 'ILMN_1651346']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_1725881', 'ILMN_1910180', 'ILMN_1804174', 'ILMN_1796063', 'ILMN_1811966'], 'nuID': ['rp13_p1x6D80lNLk3c', 'NEX0oqCV8.er4HVfU4', 'KyqQynMZxJcruyylEU', 'xXl7eXuF7sbPEp.KFI', '9ckqJrioiaej9_ajeQ'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['RefSeq', 'Unigene', 'RefSeq', 'RefSeq', 'RefSeq'], 'Search_Key': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'Transcript': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'ILMN_Gene': ['LOC23117', 'HS.575038', 'FCGR2B', 'TRIM44', 'LOC653895'], 'Source_Reference_ID': ['XM_933824.1', 'Hs.575038', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'RefSeq_ID': ['XM_933824.1', nan, 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Unigene_ID': [nan, 'Hs.575038', nan, nan, nan], 'Entrez_Gene_ID': [23117.0, nan, 2213.0, 54765.0, 653895.0], 'GI': [89040007.0, 10437021.0, 88952550.0, 29029528.0, 89033487.0], 'Accession': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Symbol': ['LOC23117', nan, 'FCGR2B', 'TRIM44', 'LOC653895'], 'Protein_Product': ['XP_938917.1', nan, 'XP_943944.1', 'NP_060053.2', 'XP_941472.1'], 'Array_Address_Id': [1710221.0, 5900364.0, 2480717.0, 1300239.0, 4480719.0], 'Probe_Type': ['I', 'S', 'I', 'S', 'S'], 'Probe_Start': [122.0, 1409.0, 1643.0, 2901.0, 25.0], 'SEQUENCE': ['GGCTCCTCTTTGGGCTCCTACTGGAATTTATCAGCCATCAGTGCATCTCT', 'ACACCTTCAGGAGGGAAGCCCTTATTTCTGGGTTGAACTCCCCTTCCATG', 'TAGGGGCAATAGGCTATACGCTACAGCCTAGGTGTGTAGTAGGCCACACC', 'CCTGCCTGTCTGCCTGTGACCTGTGTACGTATTACAGGCTTTAGGACCAG', 'CTAGCAGGGAGCGGTGAGGGAGAGCGGCTGGATTTCTTGCGGGATCTGCA'], 'Chromosome': ['16', nan, nan, '11', nan], 'Probe_Chr_Orientation': ['-', nan, nan, '+', nan], 'Probe_Coordinates': ['21766363-21766363:21769901-21769949', nan, nan, '35786070-35786119', nan], 'Cytoband': ['16p12.2a', nan, '1q23.3b', '11p13a', '10q11.23b'], 'Definition': ['PREDICTED: Homo sapiens KIAA0220-like protein, transcript variant 11 (LOC23117), mRNA.', 'Homo sapiens cDNA: FLJ21027 fis, clone CAE07110', 'PREDICTED: Homo sapiens Fc fragment of IgG, low affinity IIb, receptor (CD32) (FCGR2B), mRNA.', 'Homo sapiens tripartite motif-containing 44 (TRIM44), mRNA.', 'PREDICTED: Homo sapiens similar to protein geranylgeranyltransferase type I, beta subunit (LOC653895), mRNA.'], 'Ontology_Component': [nan, nan, nan, 'intracellular [goid 5622] [evidence IEA]', nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, 'zinc ion binding [goid 8270] [evidence IEA]; metal ion binding [goid 46872] [evidence IEA]', nan], 'Synonyms': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'Obsolete_Probe_Id': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'GB_ACC': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>nuID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Source</th>\n",
       "      <th>Search_Key</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>ILMN_Gene</th>\n",
       "      <th>Source_Reference_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>Unigene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Probe_Chr_Orientation</th>\n",
       "      <th>Probe_Coordinates</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Ontology_Component</th>\n",
       "      <th>Ontology_Process</th>\n",
       "      <th>Ontology_Function</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Obsolete_Probe_Id</th>\n",
       "      <th>GB_ACC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ILMN_1725881</td>\n",
       "      <td>rp13_p1x6D80lNLk3c</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_44919</td>\n",
       "      <td>ILMN_44919</td>\n",
       "      <td>LOC23117</td>\n",
       "      <td>XM_933824.1</td>\n",
       "      <td>XM_933824.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>21766363-21766363:21769901-21769949</td>\n",
       "      <td>16p12.2a</td>\n",
       "      <td>PREDICTED: Homo sapiens KIAA0220-like protein,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XM_933824.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ILMN_1910180</td>\n",
       "      <td>NEX0oqCV8.er4HVfU4</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Unigene</td>\n",
       "      <td>ILMN_127219</td>\n",
       "      <td>ILMN_127219</td>\n",
       "      <td>HS.575038</td>\n",
       "      <td>Hs.575038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hs.575038</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens cDNA: FLJ21027 fis, clone CAE07110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AK024680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ILMN_1804174</td>\n",
       "      <td>KyqQynMZxJcruyylEU</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_139282</td>\n",
       "      <td>ILMN_139282</td>\n",
       "      <td>FCGR2B</td>\n",
       "      <td>XM_938851.1</td>\n",
       "      <td>XM_938851.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1q23.3b</td>\n",
       "      <td>PREDICTED: Homo sapiens Fc fragment of IgG, lo...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XM_938851.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ILMN_1796063</td>\n",
       "      <td>xXl7eXuF7sbPEp.KFI</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_5006</td>\n",
       "      <td>ILMN_5006</td>\n",
       "      <td>TRIM44</td>\n",
       "      <td>NM_017583.3</td>\n",
       "      <td>NM_017583.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>+</td>\n",
       "      <td>35786070-35786119</td>\n",
       "      <td>11p13a</td>\n",
       "      <td>Homo sapiens tripartite motif-containing 44 (T...</td>\n",
       "      <td>intracellular [goid 5622] [evidence IEA]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zinc ion binding [goid 8270] [evidence IEA]; m...</td>\n",
       "      <td>MGC3490; MC7; HSA249128; DIPB</td>\n",
       "      <td>MGC3490; MC7; HSA249128; DIPB</td>\n",
       "      <td>NM_017583.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ILMN_1811966</td>\n",
       "      <td>9ckqJrioiaej9_ajeQ</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>ILMN_38756</td>\n",
       "      <td>ILMN_38756</td>\n",
       "      <td>LOC653895</td>\n",
       "      <td>XM_936379.1</td>\n",
       "      <td>XM_936379.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10q11.23b</td>\n",
       "      <td>PREDICTED: Homo sapiens similar to protein ger...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XM_936379.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630979</th>\n",
       "      <td>ILMN_2371169</td>\n",
       "      <td>11.14937973</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630980</th>\n",
       "      <td>ILMN_1701875</td>\n",
       "      <td>12.68724918</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630981</th>\n",
       "      <td>ILMN_1786396</td>\n",
       "      <td>9.097121239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630982</th>\n",
       "      <td>ILMN_1653618</td>\n",
       "      <td>7.608833313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630983</th>\n",
       "      <td>ILMN_2137536</td>\n",
       "      <td>7.301597118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>630984 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID                nuID       Species   Source   Search_Key  \\\n",
       "0       ILMN_1725881  rp13_p1x6D80lNLk3c  Homo sapiens   RefSeq   ILMN_44919   \n",
       "1       ILMN_1910180  NEX0oqCV8.er4HVfU4  Homo sapiens  Unigene  ILMN_127219   \n",
       "2       ILMN_1804174  KyqQynMZxJcruyylEU  Homo sapiens   RefSeq  ILMN_139282   \n",
       "3       ILMN_1796063  xXl7eXuF7sbPEp.KFI  Homo sapiens   RefSeq    ILMN_5006   \n",
       "4       ILMN_1811966  9ckqJrioiaej9_ajeQ  Homo sapiens   RefSeq   ILMN_38756   \n",
       "...              ...                 ...           ...      ...          ...   \n",
       "630979  ILMN_2371169         11.14937973           NaN      NaN          NaN   \n",
       "630980  ILMN_1701875         12.68724918           NaN      NaN          NaN   \n",
       "630981  ILMN_1786396         9.097121239           NaN      NaN          NaN   \n",
       "630982  ILMN_1653618         7.608833313           NaN      NaN          NaN   \n",
       "630983  ILMN_2137536         7.301597118           NaN      NaN          NaN   \n",
       "\n",
       "         Transcript  ILMN_Gene Source_Reference_ID    RefSeq_ID Unigene_ID  \\\n",
       "0        ILMN_44919   LOC23117         XM_933824.1  XM_933824.1        NaN   \n",
       "1       ILMN_127219  HS.575038           Hs.575038          NaN  Hs.575038   \n",
       "2       ILMN_139282     FCGR2B         XM_938851.1  XM_938851.1        NaN   \n",
       "3         ILMN_5006     TRIM44         NM_017583.3  NM_017583.3        NaN   \n",
       "4        ILMN_38756  LOC653895         XM_936379.1  XM_936379.1        NaN   \n",
       "...             ...        ...                 ...          ...        ...   \n",
       "630979          NaN        NaN                 NaN          NaN        NaN   \n",
       "630980          NaN        NaN                 NaN          NaN        NaN   \n",
       "630981          NaN        NaN                 NaN          NaN        NaN   \n",
       "630982          NaN        NaN                 NaN          NaN        NaN   \n",
       "630983          NaN        NaN                 NaN          NaN        NaN   \n",
       "\n",
       "        ...  Probe_Chr_Orientation                    Probe_Coordinates  \\\n",
       "0       ...                      -  21766363-21766363:21769901-21769949   \n",
       "1       ...                    NaN                                  NaN   \n",
       "2       ...                    NaN                                  NaN   \n",
       "3       ...                      +                    35786070-35786119   \n",
       "4       ...                    NaN                                  NaN   \n",
       "...     ...                    ...                                  ...   \n",
       "630979  ...                    NaN                                  NaN   \n",
       "630980  ...                    NaN                                  NaN   \n",
       "630981  ...                    NaN                                  NaN   \n",
       "630982  ...                    NaN                                  NaN   \n",
       "630983  ...                    NaN                                  NaN   \n",
       "\n",
       "         Cytoband                                         Definition  \\\n",
       "0        16p12.2a  PREDICTED: Homo sapiens KIAA0220-like protein,...   \n",
       "1             NaN    Homo sapiens cDNA: FLJ21027 fis, clone CAE07110   \n",
       "2         1q23.3b  PREDICTED: Homo sapiens Fc fragment of IgG, lo...   \n",
       "3          11p13a  Homo sapiens tripartite motif-containing 44 (T...   \n",
       "4       10q11.23b  PREDICTED: Homo sapiens similar to protein ger...   \n",
       "...           ...                                                ...   \n",
       "630979        NaN                                                NaN   \n",
       "630980        NaN                                                NaN   \n",
       "630981        NaN                                                NaN   \n",
       "630982        NaN                                                NaN   \n",
       "630983        NaN                                                NaN   \n",
       "\n",
       "                              Ontology_Component  Ontology_Process  \\\n",
       "0                                            NaN               NaN   \n",
       "1                                            NaN               NaN   \n",
       "2                                            NaN               NaN   \n",
       "3       intracellular [goid 5622] [evidence IEA]               NaN   \n",
       "4                                            NaN               NaN   \n",
       "...                                          ...               ...   \n",
       "630979                                       NaN               NaN   \n",
       "630980                                       NaN               NaN   \n",
       "630981                                       NaN               NaN   \n",
       "630982                                       NaN               NaN   \n",
       "630983                                       NaN               NaN   \n",
       "\n",
       "                                        Ontology_Function  \\\n",
       "0                                                     NaN   \n",
       "1                                                     NaN   \n",
       "2                                                     NaN   \n",
       "3       zinc ion binding [goid 8270] [evidence IEA]; m...   \n",
       "4                                                     NaN   \n",
       "...                                                   ...   \n",
       "630979                                                NaN   \n",
       "630980                                                NaN   \n",
       "630981                                                NaN   \n",
       "630982                                                NaN   \n",
       "630983                                                NaN   \n",
       "\n",
       "                             Synonyms              Obsolete_Probe_Id  \\\n",
       "0                                 NaN                            NaN   \n",
       "1                                 NaN                            NaN   \n",
       "2                                 NaN                            NaN   \n",
       "3       MGC3490; MC7; HSA249128; DIPB  MGC3490; MC7; HSA249128; DIPB   \n",
       "4                                 NaN                            NaN   \n",
       "...                               ...                            ...   \n",
       "630979                            NaN                            NaN   \n",
       "630980                            NaN                            NaN   \n",
       "630981                            NaN                            NaN   \n",
       "630982                            NaN                            NaN   \n",
       "630983                            NaN                            NaN   \n",
       "\n",
       "             GB_ACC  \n",
       "0       XM_933824.1  \n",
       "1          AK024680  \n",
       "2       XM_938851.1  \n",
       "3       NM_017583.3  \n",
       "4       XM_936379.1  \n",
       "...             ...  \n",
       "630979          NaN  \n",
       "630980          NaN  \n",
       "630981          NaN  \n",
       "630982          NaN  \n",
       "630983          NaN  \n",
       "\n",
       "[630984 rows x 30 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'nuID', 'Species', 'Source', 'Search_Key', 'Transcript',\n",
       "       'ILMN_Gene', 'Source_Reference_ID', 'RefSeq_ID', 'Unigene_ID',\n",
       "       'Entrez_Gene_ID', 'GI', 'Accession', 'Symbol', 'Protein_Product',\n",
       "       'Array_Address_Id', 'Probe_Type', 'Probe_Start', 'SEQUENCE',\n",
       "       'Chromosome', 'Probe_Chr_Orientation', 'Probe_Coordinates', 'Cytoband',\n",
       "       'Definition', 'Ontology_Component', 'Ontology_Process',\n",
       "       'Ontology_Function', 'Synonyms', 'Obsolete_Probe_Id', 'GB_ACC'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM616668</th>\n",
       "      <th>GSM616669</th>\n",
       "      <th>GSM616670</th>\n",
       "      <th>GSM616671</th>\n",
       "      <th>GSM616672</th>\n",
       "      <th>GSM616673</th>\n",
       "      <th>GSM616674</th>\n",
       "      <th>GSM616675</th>\n",
       "      <th>GSM616676</th>\n",
       "      <th>GSM616677</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM616690</th>\n",
       "      <th>GSM616691</th>\n",
       "      <th>GSM616692</th>\n",
       "      <th>GSM616693</th>\n",
       "      <th>GSM616694</th>\n",
       "      <th>GSM616695</th>\n",
       "      <th>GSM616696</th>\n",
       "      <th>GSM616697</th>\n",
       "      <th>GSM616698</th>\n",
       "      <th>GSM616699</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A26A1</th>\n",
       "      <td>6.964517</td>\n",
       "      <td>6.851497</td>\n",
       "      <td>6.821692</td>\n",
       "      <td>6.826913</td>\n",
       "      <td>6.788508</td>\n",
       "      <td>6.874905</td>\n",
       "      <td>6.896505</td>\n",
       "      <td>6.852490</td>\n",
       "      <td>6.828704</td>\n",
       "      <td>7.078653</td>\n",
       "      <td>...</td>\n",
       "      <td>7.063698</td>\n",
       "      <td>6.953264</td>\n",
       "      <td>7.004011</td>\n",
       "      <td>6.957736</td>\n",
       "      <td>6.984375</td>\n",
       "      <td>6.974978</td>\n",
       "      <td>6.945546</td>\n",
       "      <td>7.068300</td>\n",
       "      <td>6.788908</td>\n",
       "      <td>6.949239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>7.138664</td>\n",
       "      <td>7.288485</td>\n",
       "      <td>7.213284</td>\n",
       "      <td>7.165006</td>\n",
       "      <td>7.123239</td>\n",
       "      <td>7.024581</td>\n",
       "      <td>7.037256</td>\n",
       "      <td>6.882017</td>\n",
       "      <td>7.361308</td>\n",
       "      <td>7.100629</td>\n",
       "      <td>...</td>\n",
       "      <td>7.269105</td>\n",
       "      <td>7.292895</td>\n",
       "      <td>7.286026</td>\n",
       "      <td>7.194649</td>\n",
       "      <td>7.100249</td>\n",
       "      <td>7.406284</td>\n",
       "      <td>7.178009</td>\n",
       "      <td>7.214195</td>\n",
       "      <td>7.138752</td>\n",
       "      <td>7.372166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>7.104400</td>\n",
       "      <td>7.329364</td>\n",
       "      <td>7.213284</td>\n",
       "      <td>7.033386</td>\n",
       "      <td>7.183104</td>\n",
       "      <td>7.352597</td>\n",
       "      <td>7.134465</td>\n",
       "      <td>6.929381</td>\n",
       "      <td>7.067596</td>\n",
       "      <td>7.140382</td>\n",
       "      <td>...</td>\n",
       "      <td>7.090643</td>\n",
       "      <td>7.198502</td>\n",
       "      <td>7.192792</td>\n",
       "      <td>7.197270</td>\n",
       "      <td>7.168820</td>\n",
       "      <td>7.046652</td>\n",
       "      <td>6.951822</td>\n",
       "      <td>7.075513</td>\n",
       "      <td>7.323013</td>\n",
       "      <td>7.038501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACSL</th>\n",
       "      <td>7.033191</td>\n",
       "      <td>6.958621</td>\n",
       "      <td>6.788754</td>\n",
       "      <td>6.695595</td>\n",
       "      <td>7.003847</td>\n",
       "      <td>6.791738</td>\n",
       "      <td>6.751658</td>\n",
       "      <td>6.971259</td>\n",
       "      <td>6.715012</td>\n",
       "      <td>6.869732</td>\n",
       "      <td>...</td>\n",
       "      <td>6.825945</td>\n",
       "      <td>6.955751</td>\n",
       "      <td>6.838070</td>\n",
       "      <td>6.815473</td>\n",
       "      <td>6.796976</td>\n",
       "      <td>6.931289</td>\n",
       "      <td>6.823332</td>\n",
       "      <td>6.948809</td>\n",
       "      <td>6.959229</td>\n",
       "      <td>7.014102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADACL1</th>\n",
       "      <td>7.121072</td>\n",
       "      <td>6.905127</td>\n",
       "      <td>7.059623</td>\n",
       "      <td>7.089688</td>\n",
       "      <td>6.907916</td>\n",
       "      <td>7.101289</td>\n",
       "      <td>7.101122</td>\n",
       "      <td>7.056745</td>\n",
       "      <td>7.134234</td>\n",
       "      <td>7.301244</td>\n",
       "      <td>...</td>\n",
       "      <td>7.143976</td>\n",
       "      <td>7.120851</td>\n",
       "      <td>6.888729</td>\n",
       "      <td>7.024350</td>\n",
       "      <td>7.141639</td>\n",
       "      <td>7.094998</td>\n",
       "      <td>7.081976</td>\n",
       "      <td>7.119085</td>\n",
       "      <td>7.075375</td>\n",
       "      <td>6.990618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>7.914881</td>\n",
       "      <td>7.396230</td>\n",
       "      <td>7.706191</td>\n",
       "      <td>7.630652</td>\n",
       "      <td>7.340271</td>\n",
       "      <td>7.637811</td>\n",
       "      <td>8.126432</td>\n",
       "      <td>7.835849</td>\n",
       "      <td>7.917731</td>\n",
       "      <td>7.631298</td>\n",
       "      <td>...</td>\n",
       "      <td>7.630288</td>\n",
       "      <td>7.499424</td>\n",
       "      <td>7.369628</td>\n",
       "      <td>7.676353</td>\n",
       "      <td>8.130350</td>\n",
       "      <td>7.288772</td>\n",
       "      <td>7.660359</td>\n",
       "      <td>7.366716</td>\n",
       "      <td>7.832213</td>\n",
       "      <td>7.455215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bA16L21.2.1</th>\n",
       "      <td>7.339529</td>\n",
       "      <td>7.211838</td>\n",
       "      <td>7.196417</td>\n",
       "      <td>7.031924</td>\n",
       "      <td>7.290146</td>\n",
       "      <td>7.293930</td>\n",
       "      <td>7.433933</td>\n",
       "      <td>7.527881</td>\n",
       "      <td>7.461130</td>\n",
       "      <td>7.567906</td>\n",
       "      <td>...</td>\n",
       "      <td>7.326622</td>\n",
       "      <td>7.134547</td>\n",
       "      <td>6.991170</td>\n",
       "      <td>7.081851</td>\n",
       "      <td>7.406407</td>\n",
       "      <td>6.949899</td>\n",
       "      <td>7.117732</td>\n",
       "      <td>6.900108</td>\n",
       "      <td>7.034636</td>\n",
       "      <td>6.967118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ222E13.2</th>\n",
       "      <td>7.573959</td>\n",
       "      <td>7.669524</td>\n",
       "      <td>7.750307</td>\n",
       "      <td>8.019474</td>\n",
       "      <td>7.682048</td>\n",
       "      <td>7.785646</td>\n",
       "      <td>7.746305</td>\n",
       "      <td>7.526654</td>\n",
       "      <td>7.781176</td>\n",
       "      <td>7.297188</td>\n",
       "      <td>...</td>\n",
       "      <td>7.184054</td>\n",
       "      <td>7.523621</td>\n",
       "      <td>8.238818</td>\n",
       "      <td>7.646125</td>\n",
       "      <td>7.379656</td>\n",
       "      <td>7.904087</td>\n",
       "      <td>7.627300</td>\n",
       "      <td>7.496995</td>\n",
       "      <td>7.870841</td>\n",
       "      <td>7.516465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <td>7.623617</td>\n",
       "      <td>7.090202</td>\n",
       "      <td>7.306702</td>\n",
       "      <td>7.393552</td>\n",
       "      <td>7.412091</td>\n",
       "      <td>7.578962</td>\n",
       "      <td>7.972756</td>\n",
       "      <td>8.398911</td>\n",
       "      <td>7.306329</td>\n",
       "      <td>7.674254</td>\n",
       "      <td>...</td>\n",
       "      <td>7.639648</td>\n",
       "      <td>7.459779</td>\n",
       "      <td>7.044920</td>\n",
       "      <td>7.447530</td>\n",
       "      <td>7.994782</td>\n",
       "      <td>7.202621</td>\n",
       "      <td>7.403683</td>\n",
       "      <td>7.309569</td>\n",
       "      <td>7.457038</td>\n",
       "      <td>7.588035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psiTPTE22</th>\n",
       "      <td>6.842906</td>\n",
       "      <td>6.682851</td>\n",
       "      <td>6.647186</td>\n",
       "      <td>6.733926</td>\n",
       "      <td>6.669861</td>\n",
       "      <td>6.727560</td>\n",
       "      <td>6.725221</td>\n",
       "      <td>6.924751</td>\n",
       "      <td>6.757817</td>\n",
       "      <td>7.156370</td>\n",
       "      <td>...</td>\n",
       "      <td>6.760264</td>\n",
       "      <td>6.712102</td>\n",
       "      <td>6.775852</td>\n",
       "      <td>6.805877</td>\n",
       "      <td>6.854217</td>\n",
       "      <td>6.812975</td>\n",
       "      <td>6.735587</td>\n",
       "      <td>6.789018</td>\n",
       "      <td>6.619135</td>\n",
       "      <td>6.625743</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12935 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             GSM616668  GSM616669  GSM616670  GSM616671  GSM616672  GSM616673  \\\n",
       "Gene                                                                            \n",
       "A26A1         6.964517   6.851497   6.821692   6.826913   6.788508   6.874905   \n",
       "AAAS          7.138664   7.288485   7.213284   7.165006   7.123239   7.024581   \n",
       "AACS          7.104400   7.329364   7.213284   7.033386   7.183104   7.352597   \n",
       "AACSL         7.033191   6.958621   6.788754   6.695595   7.003847   6.791738   \n",
       "AADACL1       7.121072   6.905127   7.059623   7.089688   6.907916   7.101289   \n",
       "...                ...        ...        ...        ...        ...        ...   \n",
       "ZZZ3          7.914881   7.396230   7.706191   7.630652   7.340271   7.637811   \n",
       "bA16L21.2.1   7.339529   7.211838   7.196417   7.031924   7.290146   7.293930   \n",
       "dJ222E13.2    7.573959   7.669524   7.750307   8.019474   7.682048   7.785646   \n",
       "dJ341D10.1    7.623617   7.090202   7.306702   7.393552   7.412091   7.578962   \n",
       "psiTPTE22     6.842906   6.682851   6.647186   6.733926   6.669861   6.727560   \n",
       "\n",
       "             GSM616674  GSM616675  GSM616676  GSM616677  ...  GSM616690  \\\n",
       "Gene                                                     ...              \n",
       "A26A1         6.896505   6.852490   6.828704   7.078653  ...   7.063698   \n",
       "AAAS          7.037256   6.882017   7.361308   7.100629  ...   7.269105   \n",
       "AACS          7.134465   6.929381   7.067596   7.140382  ...   7.090643   \n",
       "AACSL         6.751658   6.971259   6.715012   6.869732  ...   6.825945   \n",
       "AADACL1       7.101122   7.056745   7.134234   7.301244  ...   7.143976   \n",
       "...                ...        ...        ...        ...  ...        ...   \n",
       "ZZZ3          8.126432   7.835849   7.917731   7.631298  ...   7.630288   \n",
       "bA16L21.2.1   7.433933   7.527881   7.461130   7.567906  ...   7.326622   \n",
       "dJ222E13.2    7.746305   7.526654   7.781176   7.297188  ...   7.184054   \n",
       "dJ341D10.1    7.972756   8.398911   7.306329   7.674254  ...   7.639648   \n",
       "psiTPTE22     6.725221   6.924751   6.757817   7.156370  ...   6.760264   \n",
       "\n",
       "             GSM616691  GSM616692  GSM616693  GSM616694  GSM616695  GSM616696  \\\n",
       "Gene                                                                            \n",
       "A26A1         6.953264   7.004011   6.957736   6.984375   6.974978   6.945546   \n",
       "AAAS          7.292895   7.286026   7.194649   7.100249   7.406284   7.178009   \n",
       "AACS          7.198502   7.192792   7.197270   7.168820   7.046652   6.951822   \n",
       "AACSL         6.955751   6.838070   6.815473   6.796976   6.931289   6.823332   \n",
       "AADACL1       7.120851   6.888729   7.024350   7.141639   7.094998   7.081976   \n",
       "...                ...        ...        ...        ...        ...        ...   \n",
       "ZZZ3          7.499424   7.369628   7.676353   8.130350   7.288772   7.660359   \n",
       "bA16L21.2.1   7.134547   6.991170   7.081851   7.406407   6.949899   7.117732   \n",
       "dJ222E13.2    7.523621   8.238818   7.646125   7.379656   7.904087   7.627300   \n",
       "dJ341D10.1    7.459779   7.044920   7.447530   7.994782   7.202621   7.403683   \n",
       "psiTPTE22     6.712102   6.775852   6.805877   6.854217   6.812975   6.735587   \n",
       "\n",
       "             GSM616697  GSM616698  GSM616699  \n",
       "Gene                                          \n",
       "A26A1         7.068300   6.788908   6.949239  \n",
       "AAAS          7.214195   7.138752   7.372166  \n",
       "AACS          7.075513   7.323013   7.038501  \n",
       "AACSL         6.948809   6.959229   7.014102  \n",
       "AADACL1       7.119085   7.075375   6.990618  \n",
       "...                ...        ...        ...  \n",
       "ZZZ3          7.366716   7.832213   7.455215  \n",
       "bA16L21.2.1   6.900108   7.034636   6.967118  \n",
       "dJ222E13.2    7.496995   7.870841   7.516465  \n",
       "dJ341D10.1    7.309569   7.457038   7.588035  \n",
       "psiTPTE22     6.789018   6.619135   6.625743  \n",
       "\n",
       "[12935 rows x 32 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM616668</th>\n",
       "      <th>GSM616669</th>\n",
       "      <th>GSM616670</th>\n",
       "      <th>GSM616671</th>\n",
       "      <th>GSM616672</th>\n",
       "      <th>GSM616673</th>\n",
       "      <th>GSM616674</th>\n",
       "      <th>GSM616675</th>\n",
       "      <th>GSM616676</th>\n",
       "      <th>GSM616677</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM616690</th>\n",
       "      <th>GSM616691</th>\n",
       "      <th>GSM616692</th>\n",
       "      <th>GSM616693</th>\n",
       "      <th>GSM616694</th>\n",
       "      <th>GSM616695</th>\n",
       "      <th>GSM616696</th>\n",
       "      <th>GSM616697</th>\n",
       "      <th>GSM616698</th>\n",
       "      <th>GSM616699</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>7.138664</td>\n",
       "      <td>7.288485</td>\n",
       "      <td>7.213284</td>\n",
       "      <td>7.165006</td>\n",
       "      <td>7.123239</td>\n",
       "      <td>7.024581</td>\n",
       "      <td>7.037256</td>\n",
       "      <td>6.882017</td>\n",
       "      <td>7.361308</td>\n",
       "      <td>7.100629</td>\n",
       "      <td>...</td>\n",
       "      <td>7.269105</td>\n",
       "      <td>7.292895</td>\n",
       "      <td>7.286026</td>\n",
       "      <td>7.194649</td>\n",
       "      <td>7.100249</td>\n",
       "      <td>7.406284</td>\n",
       "      <td>7.178009</td>\n",
       "      <td>7.214195</td>\n",
       "      <td>7.138752</td>\n",
       "      <td>7.372166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>7.104400</td>\n",
       "      <td>7.329364</td>\n",
       "      <td>7.213284</td>\n",
       "      <td>7.033386</td>\n",
       "      <td>7.183104</td>\n",
       "      <td>7.352597</td>\n",
       "      <td>7.134465</td>\n",
       "      <td>6.929381</td>\n",
       "      <td>7.067596</td>\n",
       "      <td>7.140382</td>\n",
       "      <td>...</td>\n",
       "      <td>7.090643</td>\n",
       "      <td>7.198502</td>\n",
       "      <td>7.192792</td>\n",
       "      <td>7.197270</td>\n",
       "      <td>7.168820</td>\n",
       "      <td>7.046652</td>\n",
       "      <td>6.951822</td>\n",
       "      <td>7.075513</td>\n",
       "      <td>7.323013</td>\n",
       "      <td>7.038501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>8.125071</td>\n",
       "      <td>8.129137</td>\n",
       "      <td>8.741529</td>\n",
       "      <td>9.285954</td>\n",
       "      <td>8.532619</td>\n",
       "      <td>8.234013</td>\n",
       "      <td>8.562658</td>\n",
       "      <td>8.690536</td>\n",
       "      <td>7.937707</td>\n",
       "      <td>7.986583</td>\n",
       "      <td>...</td>\n",
       "      <td>7.925247</td>\n",
       "      <td>8.789428</td>\n",
       "      <td>9.701271</td>\n",
       "      <td>9.514471</td>\n",
       "      <td>9.337640</td>\n",
       "      <td>9.529326</td>\n",
       "      <td>9.779090</td>\n",
       "      <td>9.065612</td>\n",
       "      <td>9.901291</td>\n",
       "      <td>9.068633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>7.782783</td>\n",
       "      <td>8.029534</td>\n",
       "      <td>7.667971</td>\n",
       "      <td>7.552847</td>\n",
       "      <td>7.936913</td>\n",
       "      <td>7.741029</td>\n",
       "      <td>7.611396</td>\n",
       "      <td>7.966206</td>\n",
       "      <td>7.819900</td>\n",
       "      <td>7.863311</td>\n",
       "      <td>...</td>\n",
       "      <td>7.882847</td>\n",
       "      <td>7.834193</td>\n",
       "      <td>8.028064</td>\n",
       "      <td>7.736413</td>\n",
       "      <td>7.761745</td>\n",
       "      <td>8.097668</td>\n",
       "      <td>8.044268</td>\n",
       "      <td>8.274695</td>\n",
       "      <td>7.868418</td>\n",
       "      <td>7.934589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS2</th>\n",
       "      <td>7.587197</td>\n",
       "      <td>7.969007</td>\n",
       "      <td>7.796640</td>\n",
       "      <td>7.702655</td>\n",
       "      <td>7.887355</td>\n",
       "      <td>7.795298</td>\n",
       "      <td>7.696078</td>\n",
       "      <td>7.563307</td>\n",
       "      <td>8.145167</td>\n",
       "      <td>7.659180</td>\n",
       "      <td>...</td>\n",
       "      <td>7.877119</td>\n",
       "      <td>7.877955</td>\n",
       "      <td>8.552603</td>\n",
       "      <td>8.044010</td>\n",
       "      <td>7.658999</td>\n",
       "      <td>8.102704</td>\n",
       "      <td>8.030426</td>\n",
       "      <td>8.248636</td>\n",
       "      <td>7.996058</td>\n",
       "      <td>7.947954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>7.378561</td>\n",
       "      <td>7.452878</td>\n",
       "      <td>7.461378</td>\n",
       "      <td>7.385134</td>\n",
       "      <td>7.467755</td>\n",
       "      <td>7.718804</td>\n",
       "      <td>7.296222</td>\n",
       "      <td>7.242961</td>\n",
       "      <td>7.242285</td>\n",
       "      <td>7.279282</td>\n",
       "      <td>...</td>\n",
       "      <td>7.569137</td>\n",
       "      <td>7.664972</td>\n",
       "      <td>7.229661</td>\n",
       "      <td>7.132043</td>\n",
       "      <td>7.310514</td>\n",
       "      <td>7.357571</td>\n",
       "      <td>7.227276</td>\n",
       "      <td>7.360502</td>\n",
       "      <td>7.446771</td>\n",
       "      <td>7.425161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>10.819175</td>\n",
       "      <td>11.302196</td>\n",
       "      <td>11.017737</td>\n",
       "      <td>10.724758</td>\n",
       "      <td>11.607532</td>\n",
       "      <td>11.321941</td>\n",
       "      <td>10.398137</td>\n",
       "      <td>10.223679</td>\n",
       "      <td>10.732550</td>\n",
       "      <td>11.002866</td>\n",
       "      <td>...</td>\n",
       "      <td>11.089995</td>\n",
       "      <td>11.514976</td>\n",
       "      <td>11.535714</td>\n",
       "      <td>10.719100</td>\n",
       "      <td>10.194335</td>\n",
       "      <td>11.138047</td>\n",
       "      <td>10.881106</td>\n",
       "      <td>10.925000</td>\n",
       "      <td>10.903630</td>\n",
       "      <td>11.353933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>11.578540</td>\n",
       "      <td>11.843250</td>\n",
       "      <td>11.460245</td>\n",
       "      <td>11.575239</td>\n",
       "      <td>12.079652</td>\n",
       "      <td>11.114155</td>\n",
       "      <td>10.822670</td>\n",
       "      <td>11.149511</td>\n",
       "      <td>11.332497</td>\n",
       "      <td>12.048542</td>\n",
       "      <td>...</td>\n",
       "      <td>12.249063</td>\n",
       "      <td>12.137544</td>\n",
       "      <td>11.776277</td>\n",
       "      <td>11.550189</td>\n",
       "      <td>11.181100</td>\n",
       "      <td>11.662707</td>\n",
       "      <td>11.189414</td>\n",
       "      <td>12.111736</td>\n",
       "      <td>11.342251</td>\n",
       "      <td>11.918314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>8.807526</td>\n",
       "      <td>8.832830</td>\n",
       "      <td>9.120240</td>\n",
       "      <td>8.602360</td>\n",
       "      <td>8.843867</td>\n",
       "      <td>9.354744</td>\n",
       "      <td>8.511277</td>\n",
       "      <td>8.494726</td>\n",
       "      <td>8.476937</td>\n",
       "      <td>8.467565</td>\n",
       "      <td>...</td>\n",
       "      <td>9.240523</td>\n",
       "      <td>9.445060</td>\n",
       "      <td>8.341862</td>\n",
       "      <td>8.362275</td>\n",
       "      <td>8.506950</td>\n",
       "      <td>9.161143</td>\n",
       "      <td>8.831561</td>\n",
       "      <td>8.752559</td>\n",
       "      <td>8.822759</td>\n",
       "      <td>9.097121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>7.914881</td>\n",
       "      <td>7.396230</td>\n",
       "      <td>7.706191</td>\n",
       "      <td>7.630652</td>\n",
       "      <td>7.340271</td>\n",
       "      <td>7.637811</td>\n",
       "      <td>8.126432</td>\n",
       "      <td>7.835849</td>\n",
       "      <td>7.917731</td>\n",
       "      <td>7.631298</td>\n",
       "      <td>...</td>\n",
       "      <td>7.630288</td>\n",
       "      <td>7.499424</td>\n",
       "      <td>7.369628</td>\n",
       "      <td>7.676353</td>\n",
       "      <td>8.130350</td>\n",
       "      <td>7.288772</td>\n",
       "      <td>7.660359</td>\n",
       "      <td>7.366716</td>\n",
       "      <td>7.832213</td>\n",
       "      <td>7.455215</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9560 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM616668  GSM616669  GSM616670  GSM616671  GSM616672  GSM616673  \\\n",
       "AAAS     7.138664   7.288485   7.213284   7.165006   7.123239   7.024581   \n",
       "AACS     7.104400   7.329364   7.213284   7.033386   7.183104   7.352597   \n",
       "AAK1     8.125071   8.129137   8.741529   9.285954   8.532619   8.234013   \n",
       "AAMP     7.782783   8.029534   7.667971   7.552847   7.936913   7.741029   \n",
       "AARS2    7.587197   7.969007   7.796640   7.702655   7.887355   7.795298   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZXDC     7.378561   7.452878   7.461378   7.385134   7.467755   7.718804   \n",
       "ZYG11B  10.819175  11.302196  11.017737  10.724758  11.607532  11.321941   \n",
       "ZYX     11.578540  11.843250  11.460245  11.575239  12.079652  11.114155   \n",
       "ZZEF1    8.807526   8.832830   9.120240   8.602360   8.843867   9.354744   \n",
       "ZZZ3     7.914881   7.396230   7.706191   7.630652   7.340271   7.637811   \n",
       "\n",
       "        GSM616674  GSM616675  GSM616676  GSM616677  ...  GSM616690  GSM616691  \\\n",
       "AAAS     7.037256   6.882017   7.361308   7.100629  ...   7.269105   7.292895   \n",
       "AACS     7.134465   6.929381   7.067596   7.140382  ...   7.090643   7.198502   \n",
       "AAK1     8.562658   8.690536   7.937707   7.986583  ...   7.925247   8.789428   \n",
       "AAMP     7.611396   7.966206   7.819900   7.863311  ...   7.882847   7.834193   \n",
       "AARS2    7.696078   7.563307   8.145167   7.659180  ...   7.877119   7.877955   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "ZXDC     7.296222   7.242961   7.242285   7.279282  ...   7.569137   7.664972   \n",
       "ZYG11B  10.398137  10.223679  10.732550  11.002866  ...  11.089995  11.514976   \n",
       "ZYX     10.822670  11.149511  11.332497  12.048542  ...  12.249063  12.137544   \n",
       "ZZEF1    8.511277   8.494726   8.476937   8.467565  ...   9.240523   9.445060   \n",
       "ZZZ3     8.126432   7.835849   7.917731   7.631298  ...   7.630288   7.499424   \n",
       "\n",
       "        GSM616692  GSM616693  GSM616694  GSM616695  GSM616696  GSM616697  \\\n",
       "AAAS     7.286026   7.194649   7.100249   7.406284   7.178009   7.214195   \n",
       "AACS     7.192792   7.197270   7.168820   7.046652   6.951822   7.075513   \n",
       "AAK1     9.701271   9.514471   9.337640   9.529326   9.779090   9.065612   \n",
       "AAMP     8.028064   7.736413   7.761745   8.097668   8.044268   8.274695   \n",
       "AARS2    8.552603   8.044010   7.658999   8.102704   8.030426   8.248636   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZXDC     7.229661   7.132043   7.310514   7.357571   7.227276   7.360502   \n",
       "ZYG11B  11.535714  10.719100  10.194335  11.138047  10.881106  10.925000   \n",
       "ZYX     11.776277  11.550189  11.181100  11.662707  11.189414  12.111736   \n",
       "ZZEF1    8.341862   8.362275   8.506950   9.161143   8.831561   8.752559   \n",
       "ZZZ3     7.369628   7.676353   8.130350   7.288772   7.660359   7.366716   \n",
       "\n",
       "        GSM616698  GSM616699  \n",
       "AAAS     7.138752   7.372166  \n",
       "AACS     7.323013   7.038501  \n",
       "AAK1     9.901291   9.068633  \n",
       "AAMP     7.868418   7.934589  \n",
       "AARS2    7.996058   7.947954  \n",
       "...           ...        ...  \n",
       "ZXDC     7.446771   7.425161  \n",
       "ZYG11B  10.903630  11.353933  \n",
       "ZYX     11.342251  11.918314  \n",
       "ZZEF1    8.822759   9.097121  \n",
       "ZZZ3     7.832213   7.455215  \n",
       "\n",
       "[9560 rows x 32 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>AARS2</th>\n",
       "      <th>AARSD1</th>\n",
       "      <th>AASDH</th>\n",
       "      <th>AASDHPPT</th>\n",
       "      <th>AASS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM5</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM616668</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.138664</td>\n",
       "      <td>7.104400</td>\n",
       "      <td>8.125071</td>\n",
       "      <td>7.782783</td>\n",
       "      <td>7.587197</td>\n",
       "      <td>7.960803</td>\n",
       "      <td>7.524874</td>\n",
       "      <td>7.715052</td>\n",
       "      <td>6.831030</td>\n",
       "      <td>...</td>\n",
       "      <td>7.006623</td>\n",
       "      <td>9.150701</td>\n",
       "      <td>7.994910</td>\n",
       "      <td>7.820794</td>\n",
       "      <td>6.958386</td>\n",
       "      <td>7.378561</td>\n",
       "      <td>10.819175</td>\n",
       "      <td>11.578540</td>\n",
       "      <td>8.807526</td>\n",
       "      <td>7.914881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616669</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.288485</td>\n",
       "      <td>7.329364</td>\n",
       "      <td>8.129137</td>\n",
       "      <td>8.029534</td>\n",
       "      <td>7.969007</td>\n",
       "      <td>8.500934</td>\n",
       "      <td>7.598680</td>\n",
       "      <td>7.167869</td>\n",
       "      <td>6.692567</td>\n",
       "      <td>...</td>\n",
       "      <td>6.884218</td>\n",
       "      <td>8.472918</td>\n",
       "      <td>8.275855</td>\n",
       "      <td>7.942832</td>\n",
       "      <td>6.977118</td>\n",
       "      <td>7.452878</td>\n",
       "      <td>11.302196</td>\n",
       "      <td>11.843250</td>\n",
       "      <td>8.832830</td>\n",
       "      <td>7.396230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616670</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.213284</td>\n",
       "      <td>7.213284</td>\n",
       "      <td>8.741529</td>\n",
       "      <td>7.667971</td>\n",
       "      <td>7.796640</td>\n",
       "      <td>8.418613</td>\n",
       "      <td>7.633838</td>\n",
       "      <td>7.162680</td>\n",
       "      <td>6.661228</td>\n",
       "      <td>...</td>\n",
       "      <td>7.016038</td>\n",
       "      <td>8.628788</td>\n",
       "      <td>8.373534</td>\n",
       "      <td>7.972513</td>\n",
       "      <td>7.027813</td>\n",
       "      <td>7.461378</td>\n",
       "      <td>11.017737</td>\n",
       "      <td>11.460245</td>\n",
       "      <td>9.120240</td>\n",
       "      <td>7.706191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616671</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.165006</td>\n",
       "      <td>7.033386</td>\n",
       "      <td>9.285954</td>\n",
       "      <td>7.552847</td>\n",
       "      <td>7.702655</td>\n",
       "      <td>8.364841</td>\n",
       "      <td>7.743772</td>\n",
       "      <td>7.285993</td>\n",
       "      <td>6.788480</td>\n",
       "      <td>...</td>\n",
       "      <td>7.097028</td>\n",
       "      <td>8.901087</td>\n",
       "      <td>7.959203</td>\n",
       "      <td>7.707246</td>\n",
       "      <td>6.907603</td>\n",
       "      <td>7.385134</td>\n",
       "      <td>10.724758</td>\n",
       "      <td>11.575239</td>\n",
       "      <td>8.602360</td>\n",
       "      <td>7.630652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616672</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.123239</td>\n",
       "      <td>7.183104</td>\n",
       "      <td>8.532619</td>\n",
       "      <td>7.936913</td>\n",
       "      <td>7.887355</td>\n",
       "      <td>8.279008</td>\n",
       "      <td>7.696225</td>\n",
       "      <td>7.200697</td>\n",
       "      <td>6.693215</td>\n",
       "      <td>...</td>\n",
       "      <td>6.958621</td>\n",
       "      <td>8.599793</td>\n",
       "      <td>8.179619</td>\n",
       "      <td>7.543519</td>\n",
       "      <td>6.980670</td>\n",
       "      <td>7.467755</td>\n",
       "      <td>11.607532</td>\n",
       "      <td>12.079652</td>\n",
       "      <td>8.843867</td>\n",
       "      <td>7.340271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616673</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.024581</td>\n",
       "      <td>7.352597</td>\n",
       "      <td>8.234013</td>\n",
       "      <td>7.741029</td>\n",
       "      <td>7.795298</td>\n",
       "      <td>8.448704</td>\n",
       "      <td>7.687576</td>\n",
       "      <td>7.496373</td>\n",
       "      <td>6.827851</td>\n",
       "      <td>...</td>\n",
       "      <td>7.068700</td>\n",
       "      <td>8.788205</td>\n",
       "      <td>8.206963</td>\n",
       "      <td>7.860533</td>\n",
       "      <td>7.144718</td>\n",
       "      <td>7.718804</td>\n",
       "      <td>11.321941</td>\n",
       "      <td>11.114155</td>\n",
       "      <td>9.354744</td>\n",
       "      <td>7.637811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616674</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.037256</td>\n",
       "      <td>7.134465</td>\n",
       "      <td>8.562658</td>\n",
       "      <td>7.611396</td>\n",
       "      <td>7.696078</td>\n",
       "      <td>8.136888</td>\n",
       "      <td>7.667861</td>\n",
       "      <td>7.852657</td>\n",
       "      <td>6.651165</td>\n",
       "      <td>...</td>\n",
       "      <td>7.015444</td>\n",
       "      <td>8.687778</td>\n",
       "      <td>8.248844</td>\n",
       "      <td>7.886782</td>\n",
       "      <td>6.963159</td>\n",
       "      <td>7.296222</td>\n",
       "      <td>10.398137</td>\n",
       "      <td>10.822670</td>\n",
       "      <td>8.511277</td>\n",
       "      <td>8.126432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616675</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.882017</td>\n",
       "      <td>6.929381</td>\n",
       "      <td>8.690536</td>\n",
       "      <td>7.966206</td>\n",
       "      <td>7.563307</td>\n",
       "      <td>7.777130</td>\n",
       "      <td>7.809420</td>\n",
       "      <td>7.672385</td>\n",
       "      <td>6.636449</td>\n",
       "      <td>...</td>\n",
       "      <td>6.788754</td>\n",
       "      <td>9.296139</td>\n",
       "      <td>7.683960</td>\n",
       "      <td>7.594610</td>\n",
       "      <td>6.895500</td>\n",
       "      <td>7.242961</td>\n",
       "      <td>10.223679</td>\n",
       "      <td>11.149511</td>\n",
       "      <td>8.494726</td>\n",
       "      <td>7.835849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616676</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.361308</td>\n",
       "      <td>7.067596</td>\n",
       "      <td>7.937707</td>\n",
       "      <td>7.819900</td>\n",
       "      <td>8.145167</td>\n",
       "      <td>8.277720</td>\n",
       "      <td>7.961760</td>\n",
       "      <td>7.502574</td>\n",
       "      <td>6.642665</td>\n",
       "      <td>...</td>\n",
       "      <td>6.822207</td>\n",
       "      <td>8.450900</td>\n",
       "      <td>8.254313</td>\n",
       "      <td>7.997812</td>\n",
       "      <td>7.100770</td>\n",
       "      <td>7.242285</td>\n",
       "      <td>10.732550</td>\n",
       "      <td>11.332497</td>\n",
       "      <td>8.476937</td>\n",
       "      <td>7.917731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616677</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.100629</td>\n",
       "      <td>7.140382</td>\n",
       "      <td>7.986583</td>\n",
       "      <td>7.863311</td>\n",
       "      <td>7.659180</td>\n",
       "      <td>7.916114</td>\n",
       "      <td>7.934813</td>\n",
       "      <td>7.337132</td>\n",
       "      <td>6.701745</td>\n",
       "      <td>...</td>\n",
       "      <td>6.763535</td>\n",
       "      <td>8.844864</td>\n",
       "      <td>8.011358</td>\n",
       "      <td>7.779078</td>\n",
       "      <td>6.895423</td>\n",
       "      <td>7.279282</td>\n",
       "      <td>11.002866</td>\n",
       "      <td>12.048542</td>\n",
       "      <td>8.467565</td>\n",
       "      <td>7.631298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616678</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.433545</td>\n",
       "      <td>7.325636</td>\n",
       "      <td>7.156370</td>\n",
       "      <td>7.847923</td>\n",
       "      <td>7.874895</td>\n",
       "      <td>8.270198</td>\n",
       "      <td>7.754902</td>\n",
       "      <td>7.262662</td>\n",
       "      <td>6.749849</td>\n",
       "      <td>...</td>\n",
       "      <td>6.852582</td>\n",
       "      <td>8.665009</td>\n",
       "      <td>8.071173</td>\n",
       "      <td>7.626564</td>\n",
       "      <td>6.855968</td>\n",
       "      <td>7.526840</td>\n",
       "      <td>10.964879</td>\n",
       "      <td>11.624672</td>\n",
       "      <td>8.741759</td>\n",
       "      <td>7.452268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616679</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.955796</td>\n",
       "      <td>7.007436</td>\n",
       "      <td>8.510966</td>\n",
       "      <td>7.840447</td>\n",
       "      <td>7.875302</td>\n",
       "      <td>7.943948</td>\n",
       "      <td>7.811047</td>\n",
       "      <td>7.635535</td>\n",
       "      <td>6.766397</td>\n",
       "      <td>...</td>\n",
       "      <td>6.870973</td>\n",
       "      <td>9.312937</td>\n",
       "      <td>7.881413</td>\n",
       "      <td>7.821149</td>\n",
       "      <td>6.801225</td>\n",
       "      <td>7.150276</td>\n",
       "      <td>9.929087</td>\n",
       "      <td>11.363516</td>\n",
       "      <td>8.330200</td>\n",
       "      <td>7.920368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616680</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.264797</td>\n",
       "      <td>7.186314</td>\n",
       "      <td>8.823807</td>\n",
       "      <td>8.015212</td>\n",
       "      <td>7.754813</td>\n",
       "      <td>8.418467</td>\n",
       "      <td>7.861661</td>\n",
       "      <td>7.409879</td>\n",
       "      <td>6.669498</td>\n",
       "      <td>...</td>\n",
       "      <td>6.994229</td>\n",
       "      <td>9.231085</td>\n",
       "      <td>8.419835</td>\n",
       "      <td>7.843686</td>\n",
       "      <td>6.892385</td>\n",
       "      <td>7.348244</td>\n",
       "      <td>10.702005</td>\n",
       "      <td>11.503050</td>\n",
       "      <td>8.636081</td>\n",
       "      <td>7.840977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616681</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.011929</td>\n",
       "      <td>7.241685</td>\n",
       "      <td>7.256524</td>\n",
       "      <td>7.730865</td>\n",
       "      <td>8.044656</td>\n",
       "      <td>7.931125</td>\n",
       "      <td>7.574689</td>\n",
       "      <td>7.351912</td>\n",
       "      <td>6.759697</td>\n",
       "      <td>...</td>\n",
       "      <td>6.798184</td>\n",
       "      <td>9.089691</td>\n",
       "      <td>8.256412</td>\n",
       "      <td>7.746595</td>\n",
       "      <td>6.862379</td>\n",
       "      <td>7.736768</td>\n",
       "      <td>11.112394</td>\n",
       "      <td>12.030608</td>\n",
       "      <td>8.911001</td>\n",
       "      <td>7.490831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616682</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.778103</td>\n",
       "      <td>6.972358</td>\n",
       "      <td>8.841487</td>\n",
       "      <td>7.643978</td>\n",
       "      <td>7.973175</td>\n",
       "      <td>8.140036</td>\n",
       "      <td>8.048448</td>\n",
       "      <td>7.582633</td>\n",
       "      <td>6.656184</td>\n",
       "      <td>...</td>\n",
       "      <td>6.959788</td>\n",
       "      <td>8.656655</td>\n",
       "      <td>7.710120</td>\n",
       "      <td>7.778952</td>\n",
       "      <td>7.240710</td>\n",
       "      <td>7.293231</td>\n",
       "      <td>10.322162</td>\n",
       "      <td>10.921897</td>\n",
       "      <td>8.460849</td>\n",
       "      <td>7.943154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616683</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.148230</td>\n",
       "      <td>7.183356</td>\n",
       "      <td>8.644526</td>\n",
       "      <td>7.595939</td>\n",
       "      <td>8.035764</td>\n",
       "      <td>8.239726</td>\n",
       "      <td>7.709988</td>\n",
       "      <td>7.313781</td>\n",
       "      <td>6.608404</td>\n",
       "      <td>...</td>\n",
       "      <td>7.057981</td>\n",
       "      <td>8.662843</td>\n",
       "      <td>8.196036</td>\n",
       "      <td>7.854103</td>\n",
       "      <td>6.938669</td>\n",
       "      <td>7.701684</td>\n",
       "      <td>11.063707</td>\n",
       "      <td>11.959617</td>\n",
       "      <td>9.217880</td>\n",
       "      <td>7.426260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616684</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.135695</td>\n",
       "      <td>7.261538</td>\n",
       "      <td>8.252893</td>\n",
       "      <td>7.997928</td>\n",
       "      <td>7.793848</td>\n",
       "      <td>8.064981</td>\n",
       "      <td>7.824405</td>\n",
       "      <td>7.615813</td>\n",
       "      <td>6.656727</td>\n",
       "      <td>...</td>\n",
       "      <td>7.022147</td>\n",
       "      <td>8.973630</td>\n",
       "      <td>8.182016</td>\n",
       "      <td>8.000066</td>\n",
       "      <td>7.011783</td>\n",
       "      <td>7.361456</td>\n",
       "      <td>10.867247</td>\n",
       "      <td>11.740149</td>\n",
       "      <td>8.916437</td>\n",
       "      <td>8.034601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616685</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.072314</td>\n",
       "      <td>7.364706</td>\n",
       "      <td>9.016638</td>\n",
       "      <td>7.668506</td>\n",
       "      <td>7.828060</td>\n",
       "      <td>8.517038</td>\n",
       "      <td>7.805102</td>\n",
       "      <td>7.426738</td>\n",
       "      <td>6.757888</td>\n",
       "      <td>...</td>\n",
       "      <td>6.968482</td>\n",
       "      <td>8.901419</td>\n",
       "      <td>8.301942</td>\n",
       "      <td>7.914487</td>\n",
       "      <td>7.034199</td>\n",
       "      <td>7.543502</td>\n",
       "      <td>11.061780</td>\n",
       "      <td>11.380827</td>\n",
       "      <td>8.783038</td>\n",
       "      <td>7.886784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616686</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.274944</td>\n",
       "      <td>7.145441</td>\n",
       "      <td>9.210965</td>\n",
       "      <td>7.856079</td>\n",
       "      <td>7.805801</td>\n",
       "      <td>8.156436</td>\n",
       "      <td>7.692396</td>\n",
       "      <td>7.428685</td>\n",
       "      <td>6.721951</td>\n",
       "      <td>...</td>\n",
       "      <td>6.912743</td>\n",
       "      <td>8.694026</td>\n",
       "      <td>8.052051</td>\n",
       "      <td>7.598999</td>\n",
       "      <td>7.026648</td>\n",
       "      <td>7.283007</td>\n",
       "      <td>11.269282</td>\n",
       "      <td>11.633214</td>\n",
       "      <td>8.954117</td>\n",
       "      <td>7.570203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616687</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.321197</td>\n",
       "      <td>7.072276</td>\n",
       "      <td>8.772198</td>\n",
       "      <td>7.797028</td>\n",
       "      <td>7.866662</td>\n",
       "      <td>8.254165</td>\n",
       "      <td>7.658000</td>\n",
       "      <td>7.142033</td>\n",
       "      <td>6.722988</td>\n",
       "      <td>...</td>\n",
       "      <td>6.944767</td>\n",
       "      <td>8.627466</td>\n",
       "      <td>8.146106</td>\n",
       "      <td>7.370667</td>\n",
       "      <td>6.896324</td>\n",
       "      <td>7.198426</td>\n",
       "      <td>11.074591</td>\n",
       "      <td>11.693334</td>\n",
       "      <td>8.878441</td>\n",
       "      <td>7.380432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616688</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.162466</td>\n",
       "      <td>7.100596</td>\n",
       "      <td>9.461217</td>\n",
       "      <td>7.770777</td>\n",
       "      <td>7.619943</td>\n",
       "      <td>8.375994</td>\n",
       "      <td>8.158908</td>\n",
       "      <td>7.818868</td>\n",
       "      <td>6.708637</td>\n",
       "      <td>...</td>\n",
       "      <td>7.037587</td>\n",
       "      <td>9.108432</td>\n",
       "      <td>8.319611</td>\n",
       "      <td>8.015070</td>\n",
       "      <td>7.075121</td>\n",
       "      <td>7.225628</td>\n",
       "      <td>10.317902</td>\n",
       "      <td>10.995097</td>\n",
       "      <td>8.578115</td>\n",
       "      <td>8.202994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616689</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.084875</td>\n",
       "      <td>7.218714</td>\n",
       "      <td>8.289656</td>\n",
       "      <td>7.575846</td>\n",
       "      <td>7.850054</td>\n",
       "      <td>8.194516</td>\n",
       "      <td>7.915998</td>\n",
       "      <td>7.938638</td>\n",
       "      <td>6.819145</td>\n",
       "      <td>...</td>\n",
       "      <td>7.071660</td>\n",
       "      <td>9.558983</td>\n",
       "      <td>8.084423</td>\n",
       "      <td>8.030290</td>\n",
       "      <td>7.050234</td>\n",
       "      <td>7.545090</td>\n",
       "      <td>10.545790</td>\n",
       "      <td>11.061608</td>\n",
       "      <td>8.771998</td>\n",
       "      <td>8.183659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616690</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.269105</td>\n",
       "      <td>7.090643</td>\n",
       "      <td>7.925247</td>\n",
       "      <td>7.882847</td>\n",
       "      <td>7.877119</td>\n",
       "      <td>8.068892</td>\n",
       "      <td>7.868640</td>\n",
       "      <td>7.366257</td>\n",
       "      <td>6.685823</td>\n",
       "      <td>...</td>\n",
       "      <td>7.255905</td>\n",
       "      <td>9.020533</td>\n",
       "      <td>8.244513</td>\n",
       "      <td>7.767542</td>\n",
       "      <td>7.061382</td>\n",
       "      <td>7.569137</td>\n",
       "      <td>11.089995</td>\n",
       "      <td>12.249063</td>\n",
       "      <td>9.240523</td>\n",
       "      <td>7.630288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616691</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.292895</td>\n",
       "      <td>7.198502</td>\n",
       "      <td>8.789428</td>\n",
       "      <td>7.834193</td>\n",
       "      <td>7.877955</td>\n",
       "      <td>7.924617</td>\n",
       "      <td>7.567906</td>\n",
       "      <td>7.267537</td>\n",
       "      <td>6.631395</td>\n",
       "      <td>...</td>\n",
       "      <td>6.904849</td>\n",
       "      <td>8.914196</td>\n",
       "      <td>7.999662</td>\n",
       "      <td>7.602739</td>\n",
       "      <td>6.815040</td>\n",
       "      <td>7.664972</td>\n",
       "      <td>11.514976</td>\n",
       "      <td>12.137544</td>\n",
       "      <td>9.445060</td>\n",
       "      <td>7.499424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616692</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.286026</td>\n",
       "      <td>7.192792</td>\n",
       "      <td>9.701271</td>\n",
       "      <td>8.028064</td>\n",
       "      <td>8.552603</td>\n",
       "      <td>8.553391</td>\n",
       "      <td>7.700584</td>\n",
       "      <td>7.027882</td>\n",
       "      <td>6.825686</td>\n",
       "      <td>...</td>\n",
       "      <td>6.898326</td>\n",
       "      <td>8.721678</td>\n",
       "      <td>8.103003</td>\n",
       "      <td>7.487344</td>\n",
       "      <td>6.796976</td>\n",
       "      <td>7.229661</td>\n",
       "      <td>11.535714</td>\n",
       "      <td>11.776277</td>\n",
       "      <td>8.341862</td>\n",
       "      <td>7.369628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616693</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.194649</td>\n",
       "      <td>7.197270</td>\n",
       "      <td>9.514471</td>\n",
       "      <td>7.736413</td>\n",
       "      <td>8.044010</td>\n",
       "      <td>8.314438</td>\n",
       "      <td>7.837830</td>\n",
       "      <td>7.497466</td>\n",
       "      <td>6.887161</td>\n",
       "      <td>...</td>\n",
       "      <td>6.875233</td>\n",
       "      <td>8.797479</td>\n",
       "      <td>8.111587</td>\n",
       "      <td>7.807005</td>\n",
       "      <td>6.916682</td>\n",
       "      <td>7.132043</td>\n",
       "      <td>10.719100</td>\n",
       "      <td>11.550189</td>\n",
       "      <td>8.362275</td>\n",
       "      <td>7.676353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616694</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.100249</td>\n",
       "      <td>7.168820</td>\n",
       "      <td>9.337640</td>\n",
       "      <td>7.761745</td>\n",
       "      <td>7.658999</td>\n",
       "      <td>8.056938</td>\n",
       "      <td>7.886598</td>\n",
       "      <td>7.979699</td>\n",
       "      <td>6.740378</td>\n",
       "      <td>...</td>\n",
       "      <td>6.799271</td>\n",
       "      <td>9.128781</td>\n",
       "      <td>8.095874</td>\n",
       "      <td>7.872118</td>\n",
       "      <td>7.127338</td>\n",
       "      <td>7.310514</td>\n",
       "      <td>10.194335</td>\n",
       "      <td>11.181100</td>\n",
       "      <td>8.506950</td>\n",
       "      <td>8.130350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616695</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.406284</td>\n",
       "      <td>7.046652</td>\n",
       "      <td>9.529326</td>\n",
       "      <td>8.097668</td>\n",
       "      <td>8.102704</td>\n",
       "      <td>8.307626</td>\n",
       "      <td>7.689620</td>\n",
       "      <td>7.203969</td>\n",
       "      <td>6.683050</td>\n",
       "      <td>...</td>\n",
       "      <td>6.918562</td>\n",
       "      <td>8.151184</td>\n",
       "      <td>8.099998</td>\n",
       "      <td>7.315849</td>\n",
       "      <td>6.902948</td>\n",
       "      <td>7.357571</td>\n",
       "      <td>11.138047</td>\n",
       "      <td>11.662707</td>\n",
       "      <td>9.161143</td>\n",
       "      <td>7.288772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616696</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.178009</td>\n",
       "      <td>6.951822</td>\n",
       "      <td>9.779090</td>\n",
       "      <td>8.044268</td>\n",
       "      <td>8.030426</td>\n",
       "      <td>8.088863</td>\n",
       "      <td>7.703003</td>\n",
       "      <td>7.655620</td>\n",
       "      <td>6.691031</td>\n",
       "      <td>...</td>\n",
       "      <td>6.962786</td>\n",
       "      <td>8.594065</td>\n",
       "      <td>8.183453</td>\n",
       "      <td>7.623725</td>\n",
       "      <td>7.031571</td>\n",
       "      <td>7.227276</td>\n",
       "      <td>10.881106</td>\n",
       "      <td>11.189414</td>\n",
       "      <td>8.831561</td>\n",
       "      <td>7.660359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616697</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.214195</td>\n",
       "      <td>7.075513</td>\n",
       "      <td>9.065612</td>\n",
       "      <td>8.274695</td>\n",
       "      <td>8.248636</td>\n",
       "      <td>8.469588</td>\n",
       "      <td>7.385751</td>\n",
       "      <td>7.187151</td>\n",
       "      <td>6.580701</td>\n",
       "      <td>...</td>\n",
       "      <td>6.925861</td>\n",
       "      <td>8.240003</td>\n",
       "      <td>8.069945</td>\n",
       "      <td>7.294838</td>\n",
       "      <td>6.973293</td>\n",
       "      <td>7.360502</td>\n",
       "      <td>10.925000</td>\n",
       "      <td>12.111736</td>\n",
       "      <td>8.752559</td>\n",
       "      <td>7.366716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616698</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.138752</td>\n",
       "      <td>7.323013</td>\n",
       "      <td>9.901291</td>\n",
       "      <td>7.868418</td>\n",
       "      <td>7.996058</td>\n",
       "      <td>8.215244</td>\n",
       "      <td>7.648363</td>\n",
       "      <td>7.532516</td>\n",
       "      <td>7.426879</td>\n",
       "      <td>...</td>\n",
       "      <td>6.893475</td>\n",
       "      <td>8.363867</td>\n",
       "      <td>8.420900</td>\n",
       "      <td>7.618671</td>\n",
       "      <td>6.904607</td>\n",
       "      <td>7.446771</td>\n",
       "      <td>10.903630</td>\n",
       "      <td>11.342251</td>\n",
       "      <td>8.822759</td>\n",
       "      <td>7.832213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM616699</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.372166</td>\n",
       "      <td>7.038501</td>\n",
       "      <td>9.068633</td>\n",
       "      <td>7.934589</td>\n",
       "      <td>7.947954</td>\n",
       "      <td>7.963468</td>\n",
       "      <td>7.459534</td>\n",
       "      <td>7.292837</td>\n",
       "      <td>6.629463</td>\n",
       "      <td>...</td>\n",
       "      <td>6.888931</td>\n",
       "      <td>8.718143</td>\n",
       "      <td>8.059211</td>\n",
       "      <td>7.377203</td>\n",
       "      <td>6.843797</td>\n",
       "      <td>7.425161</td>\n",
       "      <td>11.353933</td>\n",
       "      <td>11.918314</td>\n",
       "      <td>9.097121</td>\n",
       "      <td>7.455215</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32 rows Ã— 9561 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Ankylosing Spondylitis      AAAS      AACS      AAK1      AAMP  \\\n",
       "GSM616668                     1.0  7.138664  7.104400  8.125071  7.782783   \n",
       "GSM616669                     1.0  7.288485  7.329364  8.129137  8.029534   \n",
       "GSM616670                     1.0  7.213284  7.213284  8.741529  7.667971   \n",
       "GSM616671                     1.0  7.165006  7.033386  9.285954  7.552847   \n",
       "GSM616672                     1.0  7.123239  7.183104  8.532619  7.936913   \n",
       "GSM616673                     1.0  7.024581  7.352597  8.234013  7.741029   \n",
       "GSM616674                     1.0  7.037256  7.134465  8.562658  7.611396   \n",
       "GSM616675                     1.0  6.882017  6.929381  8.690536  7.966206   \n",
       "GSM616676                     1.0  7.361308  7.067596  7.937707  7.819900   \n",
       "GSM616677                     1.0  7.100629  7.140382  7.986583  7.863311   \n",
       "GSM616678                     1.0  7.433545  7.325636  7.156370  7.847923   \n",
       "GSM616679                     1.0  6.955796  7.007436  8.510966  7.840447   \n",
       "GSM616680                     1.0  7.264797  7.186314  8.823807  8.015212   \n",
       "GSM616681                     1.0  7.011929  7.241685  7.256524  7.730865   \n",
       "GSM616682                     1.0  6.778103  6.972358  8.841487  7.643978   \n",
       "GSM616683                     1.0  7.148230  7.183356  8.644526  7.595939   \n",
       "GSM616684                     0.0  7.135695  7.261538  8.252893  7.997928   \n",
       "GSM616685                     0.0  7.072314  7.364706  9.016638  7.668506   \n",
       "GSM616686                     0.0  7.274944  7.145441  9.210965  7.856079   \n",
       "GSM616687                     0.0  7.321197  7.072276  8.772198  7.797028   \n",
       "GSM616688                     0.0  7.162466  7.100596  9.461217  7.770777   \n",
       "GSM616689                     0.0  7.084875  7.218714  8.289656  7.575846   \n",
       "GSM616690                     0.0  7.269105  7.090643  7.925247  7.882847   \n",
       "GSM616691                     0.0  7.292895  7.198502  8.789428  7.834193   \n",
       "GSM616692                     0.0  7.286026  7.192792  9.701271  8.028064   \n",
       "GSM616693                     0.0  7.194649  7.197270  9.514471  7.736413   \n",
       "GSM616694                     0.0  7.100249  7.168820  9.337640  7.761745   \n",
       "GSM616695                     0.0  7.406284  7.046652  9.529326  8.097668   \n",
       "GSM616696                     0.0  7.178009  6.951822  9.779090  8.044268   \n",
       "GSM616697                     0.0  7.214195  7.075513  9.065612  8.274695   \n",
       "GSM616698                     0.0  7.138752  7.323013  9.901291  7.868418   \n",
       "GSM616699                     0.0  7.372166  7.038501  9.068633  7.934589   \n",
       "\n",
       "              AARS2    AARSD1     AASDH  AASDHPPT      AASS  ...    ZSWIM5  \\\n",
       "GSM616668  7.587197  7.960803  7.524874  7.715052  6.831030  ...  7.006623   \n",
       "GSM616669  7.969007  8.500934  7.598680  7.167869  6.692567  ...  6.884218   \n",
       "GSM616670  7.796640  8.418613  7.633838  7.162680  6.661228  ...  7.016038   \n",
       "GSM616671  7.702655  8.364841  7.743772  7.285993  6.788480  ...  7.097028   \n",
       "GSM616672  7.887355  8.279008  7.696225  7.200697  6.693215  ...  6.958621   \n",
       "GSM616673  7.795298  8.448704  7.687576  7.496373  6.827851  ...  7.068700   \n",
       "GSM616674  7.696078  8.136888  7.667861  7.852657  6.651165  ...  7.015444   \n",
       "GSM616675  7.563307  7.777130  7.809420  7.672385  6.636449  ...  6.788754   \n",
       "GSM616676  8.145167  8.277720  7.961760  7.502574  6.642665  ...  6.822207   \n",
       "GSM616677  7.659180  7.916114  7.934813  7.337132  6.701745  ...  6.763535   \n",
       "GSM616678  7.874895  8.270198  7.754902  7.262662  6.749849  ...  6.852582   \n",
       "GSM616679  7.875302  7.943948  7.811047  7.635535  6.766397  ...  6.870973   \n",
       "GSM616680  7.754813  8.418467  7.861661  7.409879  6.669498  ...  6.994229   \n",
       "GSM616681  8.044656  7.931125  7.574689  7.351912  6.759697  ...  6.798184   \n",
       "GSM616682  7.973175  8.140036  8.048448  7.582633  6.656184  ...  6.959788   \n",
       "GSM616683  8.035764  8.239726  7.709988  7.313781  6.608404  ...  7.057981   \n",
       "GSM616684  7.793848  8.064981  7.824405  7.615813  6.656727  ...  7.022147   \n",
       "GSM616685  7.828060  8.517038  7.805102  7.426738  6.757888  ...  6.968482   \n",
       "GSM616686  7.805801  8.156436  7.692396  7.428685  6.721951  ...  6.912743   \n",
       "GSM616687  7.866662  8.254165  7.658000  7.142033  6.722988  ...  6.944767   \n",
       "GSM616688  7.619943  8.375994  8.158908  7.818868  6.708637  ...  7.037587   \n",
       "GSM616689  7.850054  8.194516  7.915998  7.938638  6.819145  ...  7.071660   \n",
       "GSM616690  7.877119  8.068892  7.868640  7.366257  6.685823  ...  7.255905   \n",
       "GSM616691  7.877955  7.924617  7.567906  7.267537  6.631395  ...  6.904849   \n",
       "GSM616692  8.552603  8.553391  7.700584  7.027882  6.825686  ...  6.898326   \n",
       "GSM616693  8.044010  8.314438  7.837830  7.497466  6.887161  ...  6.875233   \n",
       "GSM616694  7.658999  8.056938  7.886598  7.979699  6.740378  ...  6.799271   \n",
       "GSM616695  8.102704  8.307626  7.689620  7.203969  6.683050  ...  6.918562   \n",
       "GSM616696  8.030426  8.088863  7.703003  7.655620  6.691031  ...  6.962786   \n",
       "GSM616697  8.248636  8.469588  7.385751  7.187151  6.580701  ...  6.925861   \n",
       "GSM616698  7.996058  8.215244  7.648363  7.532516  7.426879  ...  6.893475   \n",
       "GSM616699  7.947954  7.963468  7.459534  7.292837  6.629463  ...  6.888931   \n",
       "\n",
       "             ZSWIM6      ZW10    ZWILCH      ZXDB      ZXDC     ZYG11B  \\\n",
       "GSM616668  9.150701  7.994910  7.820794  6.958386  7.378561  10.819175   \n",
       "GSM616669  8.472918  8.275855  7.942832  6.977118  7.452878  11.302196   \n",
       "GSM616670  8.628788  8.373534  7.972513  7.027813  7.461378  11.017737   \n",
       "GSM616671  8.901087  7.959203  7.707246  6.907603  7.385134  10.724758   \n",
       "GSM616672  8.599793  8.179619  7.543519  6.980670  7.467755  11.607532   \n",
       "GSM616673  8.788205  8.206963  7.860533  7.144718  7.718804  11.321941   \n",
       "GSM616674  8.687778  8.248844  7.886782  6.963159  7.296222  10.398137   \n",
       "GSM616675  9.296139  7.683960  7.594610  6.895500  7.242961  10.223679   \n",
       "GSM616676  8.450900  8.254313  7.997812  7.100770  7.242285  10.732550   \n",
       "GSM616677  8.844864  8.011358  7.779078  6.895423  7.279282  11.002866   \n",
       "GSM616678  8.665009  8.071173  7.626564  6.855968  7.526840  10.964879   \n",
       "GSM616679  9.312937  7.881413  7.821149  6.801225  7.150276   9.929087   \n",
       "GSM616680  9.231085  8.419835  7.843686  6.892385  7.348244  10.702005   \n",
       "GSM616681  9.089691  8.256412  7.746595  6.862379  7.736768  11.112394   \n",
       "GSM616682  8.656655  7.710120  7.778952  7.240710  7.293231  10.322162   \n",
       "GSM616683  8.662843  8.196036  7.854103  6.938669  7.701684  11.063707   \n",
       "GSM616684  8.973630  8.182016  8.000066  7.011783  7.361456  10.867247   \n",
       "GSM616685  8.901419  8.301942  7.914487  7.034199  7.543502  11.061780   \n",
       "GSM616686  8.694026  8.052051  7.598999  7.026648  7.283007  11.269282   \n",
       "GSM616687  8.627466  8.146106  7.370667  6.896324  7.198426  11.074591   \n",
       "GSM616688  9.108432  8.319611  8.015070  7.075121  7.225628  10.317902   \n",
       "GSM616689  9.558983  8.084423  8.030290  7.050234  7.545090  10.545790   \n",
       "GSM616690  9.020533  8.244513  7.767542  7.061382  7.569137  11.089995   \n",
       "GSM616691  8.914196  7.999662  7.602739  6.815040  7.664972  11.514976   \n",
       "GSM616692  8.721678  8.103003  7.487344  6.796976  7.229661  11.535714   \n",
       "GSM616693  8.797479  8.111587  7.807005  6.916682  7.132043  10.719100   \n",
       "GSM616694  9.128781  8.095874  7.872118  7.127338  7.310514  10.194335   \n",
       "GSM616695  8.151184  8.099998  7.315849  6.902948  7.357571  11.138047   \n",
       "GSM616696  8.594065  8.183453  7.623725  7.031571  7.227276  10.881106   \n",
       "GSM616697  8.240003  8.069945  7.294838  6.973293  7.360502  10.925000   \n",
       "GSM616698  8.363867  8.420900  7.618671  6.904607  7.446771  10.903630   \n",
       "GSM616699  8.718143  8.059211  7.377203  6.843797  7.425161  11.353933   \n",
       "\n",
       "                 ZYX     ZZEF1      ZZZ3  \n",
       "GSM616668  11.578540  8.807526  7.914881  \n",
       "GSM616669  11.843250  8.832830  7.396230  \n",
       "GSM616670  11.460245  9.120240  7.706191  \n",
       "GSM616671  11.575239  8.602360  7.630652  \n",
       "GSM616672  12.079652  8.843867  7.340271  \n",
       "GSM616673  11.114155  9.354744  7.637811  \n",
       "GSM616674  10.822670  8.511277  8.126432  \n",
       "GSM616675  11.149511  8.494726  7.835849  \n",
       "GSM616676  11.332497  8.476937  7.917731  \n",
       "GSM616677  12.048542  8.467565  7.631298  \n",
       "GSM616678  11.624672  8.741759  7.452268  \n",
       "GSM616679  11.363516  8.330200  7.920368  \n",
       "GSM616680  11.503050  8.636081  7.840977  \n",
       "GSM616681  12.030608  8.911001  7.490831  \n",
       "GSM616682  10.921897  8.460849  7.943154  \n",
       "GSM616683  11.959617  9.217880  7.426260  \n",
       "GSM616684  11.740149  8.916437  8.034601  \n",
       "GSM616685  11.380827  8.783038  7.886784  \n",
       "GSM616686  11.633214  8.954117  7.570203  \n",
       "GSM616687  11.693334  8.878441  7.380432  \n",
       "GSM616688  10.995097  8.578115  8.202994  \n",
       "GSM616689  11.061608  8.771998  8.183659  \n",
       "GSM616690  12.249063  9.240523  7.630288  \n",
       "GSM616691  12.137544  9.445060  7.499424  \n",
       "GSM616692  11.776277  8.341862  7.369628  \n",
       "GSM616693  11.550189  8.362275  7.676353  \n",
       "GSM616694  11.181100  8.506950  8.130350  \n",
       "GSM616695  11.662707  9.161143  7.288772  \n",
       "GSM616696  11.189414  8.831561  7.660359  \n",
       "GSM616697  12.111736  8.752559  7.366716  \n",
       "GSM616698  11.342251  8.822759  7.832213  \n",
       "GSM616699  11.918314  9.097121  7.455215  \n",
       "\n",
       "[32 rows x 9561 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 32 samples.\n",
      "For the feature 'Ankylosing Spondylitis', the least common label is '1.0' with 16 occurrences. This represents 50.00% of the dataset.\n",
      "The distribution of the feature 'Ankylosing Spondylitis' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Sexual Dimorphism in the Th17 Signature of Ankylosing Spondylitis\"\n",
      "!Series_summary\t\"Male AS patients have an elevated Th17 cell frequency vs. female AS patients (Gracey et al, Arthritis and Rheumatology, 2015). This analysis was performed to further examine differences between male and female AS patients\"\n",
      "!Series_overall_design\t\"AS patients were compared to healthy controls (HC). For sex-specific anaylsis, three groups were compared: F-HC vs. M-HC, M-AS vs. M-HC and F-AS vs. F-HC. A one way ANOVA was performed to identify genes differentially regulated in male and female AS patients\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1902130</th>\n",
       "      <th>GSM1902131</th>\n",
       "      <th>GSM1902132</th>\n",
       "      <th>GSM1902133</th>\n",
       "      <th>GSM1902134</th>\n",
       "      <th>GSM1902135</th>\n",
       "      <th>GSM1902136</th>\n",
       "      <th>GSM1902137</th>\n",
       "      <th>GSM1902138</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1902192</th>\n",
       "      <th>GSM1902193</th>\n",
       "      <th>GSM1902194</th>\n",
       "      <th>GSM1902195</th>\n",
       "      <th>GSM1902196</th>\n",
       "      <th>GSM1902197</th>\n",
       "      <th>GSM1902198</th>\n",
       "      <th>GSM1902199</th>\n",
       "      <th>GSM1902200</th>\n",
       "      <th>GSM1902201</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>Sex: Male</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "      <td>Sex: Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age (yr): 53</td>\n",
       "      <td>age (yr): 26</td>\n",
       "      <td>age (yr): 29</td>\n",
       "      <td>age (yr): 50</td>\n",
       "      <td>age (yr): 35</td>\n",
       "      <td>age (yr): 48</td>\n",
       "      <td>age (yr): 18</td>\n",
       "      <td>age (yr): 39</td>\n",
       "      <td>age (yr): 49</td>\n",
       "      <td>...</td>\n",
       "      <td>age (yr): 27</td>\n",
       "      <td>age (yr): 37</td>\n",
       "      <td>age (yr): 42</td>\n",
       "      <td>age (yr): 63</td>\n",
       "      <td>age (yr): 61</td>\n",
       "      <td>age (yr): 20</td>\n",
       "      <td>age (yr): 31</td>\n",
       "      <td>age (yr): 25</td>\n",
       "      <td>age (yr): 29</td>\n",
       "      <td>age (yr): 65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 0</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): 1</td>\n",
       "      <td>...</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "      <td>hla-b27 (1=positive, 0=negative): unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>disease: Ankylosing Spondylitis</td>\n",
       "      <td>...</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "      <td>disease: healthy control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM1902130  \\\n",
       "0  !Sample_characteristics_ch1                            Sex: Male   \n",
       "1  !Sample_characteristics_ch1                         age (yr): 53   \n",
       "2  !Sample_characteristics_ch1  hla-b27 (1=positive, 0=negative): 1   \n",
       "3  !Sample_characteristics_ch1      disease: Ankylosing Spondylitis   \n",
       "\n",
       "                            GSM1902131                           GSM1902132  \\\n",
       "0                            Sex: Male                            Sex: Male   \n",
       "1                         age (yr): 26                         age (yr): 29   \n",
       "2  hla-b27 (1=positive, 0=negative): 1  hla-b27 (1=positive, 0=negative): 1   \n",
       "3      disease: Ankylosing Spondylitis      disease: Ankylosing Spondylitis   \n",
       "\n",
       "                            GSM1902133                           GSM1902134  \\\n",
       "0                            Sex: Male                            Sex: Male   \n",
       "1                         age (yr): 50                         age (yr): 35   \n",
       "2  hla-b27 (1=positive, 0=negative): 1  hla-b27 (1=positive, 0=negative): 1   \n",
       "3      disease: Ankylosing Spondylitis      disease: Ankylosing Spondylitis   \n",
       "\n",
       "                            GSM1902135                           GSM1902136  \\\n",
       "0                            Sex: Male                            Sex: Male   \n",
       "1                         age (yr): 48                         age (yr): 18   \n",
       "2  hla-b27 (1=positive, 0=negative): 0  hla-b27 (1=positive, 0=negative): 1   \n",
       "3      disease: Ankylosing Spondylitis      disease: Ankylosing Spondylitis   \n",
       "\n",
       "                            GSM1902137                           GSM1902138  \\\n",
       "0                            Sex: Male                            Sex: Male   \n",
       "1                         age (yr): 39                         age (yr): 49   \n",
       "2  hla-b27 (1=positive, 0=negative): 1  hla-b27 (1=positive, 0=negative): 1   \n",
       "3      disease: Ankylosing Spondylitis      disease: Ankylosing Spondylitis   \n",
       "\n",
       "   ...                                 GSM1902192  \\\n",
       "0  ...                                Sex: Female   \n",
       "1  ...                               age (yr): 27   \n",
       "2  ...  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3  ...                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902193  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 37   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902194  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 42   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902195  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 63   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902196  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 61   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902197  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 20   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902198  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 31   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902199  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 25   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902200  \\\n",
       "0                                Sex: Female   \n",
       "1                               age (yr): 29   \n",
       "2  hla-b27 (1=positive, 0=negative): unknown   \n",
       "3                   disease: healthy control   \n",
       "\n",
       "                                  GSM1902201  \n",
       "0                                Sex: Female  \n",
       "1                               age (yr): 65  \n",
       "2  hla-b27 (1=positive, 0=negative): unknown  \n",
       "3                   disease: healthy control  \n",
       "\n",
       "[4 rows x 73 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE73754\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease: Ankylosing Spondylitis',\n",
       "       'disease: healthy control'], dtype=object)"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[3]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 3\n",
    "age_row = 1\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease: Ankylosing Spondylitis':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1902130</th>\n",
       "      <th>GSM1902131</th>\n",
       "      <th>GSM1902132</th>\n",
       "      <th>GSM1902133</th>\n",
       "      <th>GSM1902134</th>\n",
       "      <th>GSM1902135</th>\n",
       "      <th>GSM1902136</th>\n",
       "      <th>GSM1902137</th>\n",
       "      <th>GSM1902138</th>\n",
       "      <th>GSM1902139</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1902192</th>\n",
       "      <th>GSM1902193</th>\n",
       "      <th>GSM1902194</th>\n",
       "      <th>GSM1902195</th>\n",
       "      <th>GSM1902196</th>\n",
       "      <th>GSM1902197</th>\n",
       "      <th>GSM1902198</th>\n",
       "      <th>GSM1902199</th>\n",
       "      <th>GSM1902200</th>\n",
       "      <th>GSM1902201</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>53</td>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>50</td>\n",
       "      <td>35</td>\n",
       "      <td>48</td>\n",
       "      <td>18</td>\n",
       "      <td>39</td>\n",
       "      <td>49</td>\n",
       "      <td>43</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>37</td>\n",
       "      <td>42</td>\n",
       "      <td>63</td>\n",
       "      <td>61</td>\n",
       "      <td>20</td>\n",
       "      <td>31</td>\n",
       "      <td>25</td>\n",
       "      <td>29</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GSM1902130  GSM1902131  GSM1902132  GSM1902133  \\\n",
       "Ankylosing Spondylitis           1           1           1           1   \n",
       "Age                             53          26          29          50   \n",
       "Gender                           0           0           0           0   \n",
       "\n",
       "                        GSM1902134  GSM1902135  GSM1902136  GSM1902137  \\\n",
       "Ankylosing Spondylitis           1           1           1           1   \n",
       "Age                             35          48          18          39   \n",
       "Gender                           0           0           0           0   \n",
       "\n",
       "                        GSM1902138  GSM1902139  ...  GSM1902192  GSM1902193  \\\n",
       "Ankylosing Spondylitis           1           1  ...           0           0   \n",
       "Age                             49          43  ...          27          37   \n",
       "Gender                           0           0  ...           1           1   \n",
       "\n",
       "                        GSM1902194  GSM1902195  GSM1902196  GSM1902197  \\\n",
       "Ankylosing Spondylitis           0           0           0           0   \n",
       "Age                             42          63          61          20   \n",
       "Gender                           1           1           1           1   \n",
       "\n",
       "                        GSM1902198  GSM1902199  GSM1902200  GSM1902201  \n",
       "Ankylosing Spondylitis           0           0           0           0  \n",
       "Age                             31          25          29          65  \n",
       "Gender                           1           1           1           1  \n",
       "\n",
       "[3 rows x 72 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1902130</th>\n",
       "      <th>GSM1902131</th>\n",
       "      <th>GSM1902132</th>\n",
       "      <th>GSM1902133</th>\n",
       "      <th>GSM1902134</th>\n",
       "      <th>GSM1902135</th>\n",
       "      <th>GSM1902136</th>\n",
       "      <th>GSM1902137</th>\n",
       "      <th>GSM1902138</th>\n",
       "      <th>GSM1902139</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1902192</th>\n",
       "      <th>GSM1902193</th>\n",
       "      <th>GSM1902194</th>\n",
       "      <th>GSM1902195</th>\n",
       "      <th>GSM1902196</th>\n",
       "      <th>GSM1902197</th>\n",
       "      <th>GSM1902198</th>\n",
       "      <th>GSM1902199</th>\n",
       "      <th>GSM1902200</th>\n",
       "      <th>GSM1902201</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>14.723105</td>\n",
       "      <td>14.974196</td>\n",
       "      <td>14.875049</td>\n",
       "      <td>14.262832</td>\n",
       "      <td>14.380508</td>\n",
       "      <td>14.737535</td>\n",
       "      <td>14.994368</td>\n",
       "      <td>14.517866</td>\n",
       "      <td>14.751615</td>\n",
       "      <td>14.937985</td>\n",
       "      <td>...</td>\n",
       "      <td>14.546172</td>\n",
       "      <td>14.609772</td>\n",
       "      <td>14.502806</td>\n",
       "      <td>14.751615</td>\n",
       "      <td>15.046889</td>\n",
       "      <td>14.751615</td>\n",
       "      <td>14.890210</td>\n",
       "      <td>14.859273</td>\n",
       "      <td>14.937985</td>\n",
       "      <td>14.455832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1343295</th>\n",
       "      <td>10.796875</td>\n",
       "      <td>10.545005</td>\n",
       "      <td>10.617579</td>\n",
       "      <td>10.663274</td>\n",
       "      <td>10.373557</td>\n",
       "      <td>11.070106</td>\n",
       "      <td>10.609179</td>\n",
       "      <td>10.614307</td>\n",
       "      <td>10.612894</td>\n",
       "      <td>10.263954</td>\n",
       "      <td>...</td>\n",
       "      <td>10.551885</td>\n",
       "      <td>10.677234</td>\n",
       "      <td>10.606567</td>\n",
       "      <td>10.688255</td>\n",
       "      <td>10.647923</td>\n",
       "      <td>10.527452</td>\n",
       "      <td>10.519320</td>\n",
       "      <td>10.395747</td>\n",
       "      <td>10.680664</td>\n",
       "      <td>10.368892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651199</th>\n",
       "      <td>6.625879</td>\n",
       "      <td>6.658650</td>\n",
       "      <td>6.614950</td>\n",
       "      <td>6.729367</td>\n",
       "      <td>6.670856</td>\n",
       "      <td>6.751337</td>\n",
       "      <td>6.658062</td>\n",
       "      <td>6.658457</td>\n",
       "      <td>6.644127</td>\n",
       "      <td>6.698253</td>\n",
       "      <td>...</td>\n",
       "      <td>6.663772</td>\n",
       "      <td>6.603572</td>\n",
       "      <td>6.649300</td>\n",
       "      <td>6.763689</td>\n",
       "      <td>6.677779</td>\n",
       "      <td>6.550622</td>\n",
       "      <td>6.693903</td>\n",
       "      <td>6.742800</td>\n",
       "      <td>6.584209</td>\n",
       "      <td>6.593124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651209</th>\n",
       "      <td>6.762562</td>\n",
       "      <td>6.775621</td>\n",
       "      <td>6.630290</td>\n",
       "      <td>6.822824</td>\n",
       "      <td>6.670856</td>\n",
       "      <td>6.713077</td>\n",
       "      <td>6.704106</td>\n",
       "      <td>6.854264</td>\n",
       "      <td>6.767462</td>\n",
       "      <td>6.659816</td>\n",
       "      <td>...</td>\n",
       "      <td>6.835929</td>\n",
       "      <td>6.729711</td>\n",
       "      <td>6.770642</td>\n",
       "      <td>6.811416</td>\n",
       "      <td>6.824817</td>\n",
       "      <td>6.736938</td>\n",
       "      <td>6.766212</td>\n",
       "      <td>6.765388</td>\n",
       "      <td>6.690326</td>\n",
       "      <td>6.719627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651210</th>\n",
       "      <td>6.800612</td>\n",
       "      <td>6.596698</td>\n",
       "      <td>6.613691</td>\n",
       "      <td>6.504107</td>\n",
       "      <td>6.708310</td>\n",
       "      <td>6.593606</td>\n",
       "      <td>6.674507</td>\n",
       "      <td>6.741474</td>\n",
       "      <td>6.683399</td>\n",
       "      <td>6.752126</td>\n",
       "      <td>...</td>\n",
       "      <td>6.612110</td>\n",
       "      <td>6.662521</td>\n",
       "      <td>6.708362</td>\n",
       "      <td>6.758281</td>\n",
       "      <td>6.690449</td>\n",
       "      <td>6.609686</td>\n",
       "      <td>6.743372</td>\n",
       "      <td>6.715278</td>\n",
       "      <td>6.601116</td>\n",
       "      <td>6.584874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311170</th>\n",
       "      <td>6.598350</td>\n",
       "      <td>6.656326</td>\n",
       "      <td>6.745738</td>\n",
       "      <td>6.566161</td>\n",
       "      <td>6.572376</td>\n",
       "      <td>6.582429</td>\n",
       "      <td>6.613186</td>\n",
       "      <td>6.614191</td>\n",
       "      <td>6.634587</td>\n",
       "      <td>6.679710</td>\n",
       "      <td>...</td>\n",
       "      <td>6.665184</td>\n",
       "      <td>6.615608</td>\n",
       "      <td>6.665901</td>\n",
       "      <td>6.591221</td>\n",
       "      <td>6.563782</td>\n",
       "      <td>6.734294</td>\n",
       "      <td>6.627239</td>\n",
       "      <td>6.496727</td>\n",
       "      <td>6.703297</td>\n",
       "      <td>6.643118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311175</th>\n",
       "      <td>6.652280</td>\n",
       "      <td>6.639231</td>\n",
       "      <td>6.695448</td>\n",
       "      <td>6.540806</td>\n",
       "      <td>6.721645</td>\n",
       "      <td>6.662629</td>\n",
       "      <td>6.662562</td>\n",
       "      <td>6.680551</td>\n",
       "      <td>6.638602</td>\n",
       "      <td>6.637241</td>\n",
       "      <td>...</td>\n",
       "      <td>6.673474</td>\n",
       "      <td>6.650218</td>\n",
       "      <td>6.685569</td>\n",
       "      <td>6.740448</td>\n",
       "      <td>6.637933</td>\n",
       "      <td>6.638491</td>\n",
       "      <td>6.666933</td>\n",
       "      <td>6.588203</td>\n",
       "      <td>6.743888</td>\n",
       "      <td>6.641553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311180</th>\n",
       "      <td>6.864999</td>\n",
       "      <td>6.839016</td>\n",
       "      <td>6.642746</td>\n",
       "      <td>6.691449</td>\n",
       "      <td>6.716093</td>\n",
       "      <td>6.657982</td>\n",
       "      <td>6.662562</td>\n",
       "      <td>6.669022</td>\n",
       "      <td>6.671518</td>\n",
       "      <td>6.704380</td>\n",
       "      <td>...</td>\n",
       "      <td>6.770264</td>\n",
       "      <td>6.613962</td>\n",
       "      <td>6.695612</td>\n",
       "      <td>6.821868</td>\n",
       "      <td>6.648760</td>\n",
       "      <td>6.698223</td>\n",
       "      <td>6.615875</td>\n",
       "      <td>6.737687</td>\n",
       "      <td>6.725904</td>\n",
       "      <td>6.682509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311185</th>\n",
       "      <td>6.632191</td>\n",
       "      <td>6.655011</td>\n",
       "      <td>6.718722</td>\n",
       "      <td>6.732387</td>\n",
       "      <td>6.545948</td>\n",
       "      <td>6.675332</td>\n",
       "      <td>6.608312</td>\n",
       "      <td>6.635070</td>\n",
       "      <td>6.621255</td>\n",
       "      <td>6.705511</td>\n",
       "      <td>...</td>\n",
       "      <td>6.742424</td>\n",
       "      <td>6.655760</td>\n",
       "      <td>6.705492</td>\n",
       "      <td>6.636093</td>\n",
       "      <td>6.659603</td>\n",
       "      <td>6.696920</td>\n",
       "      <td>6.622859</td>\n",
       "      <td>6.716543</td>\n",
       "      <td>6.626361</td>\n",
       "      <td>6.704373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311190</th>\n",
       "      <td>6.709582</td>\n",
       "      <td>6.883397</td>\n",
       "      <td>6.685806</td>\n",
       "      <td>6.922629</td>\n",
       "      <td>6.997787</td>\n",
       "      <td>6.838690</td>\n",
       "      <td>6.824616</td>\n",
       "      <td>6.937539</td>\n",
       "      <td>6.860919</td>\n",
       "      <td>6.901888</td>\n",
       "      <td>...</td>\n",
       "      <td>6.978754</td>\n",
       "      <td>7.081290</td>\n",
       "      <td>7.009742</td>\n",
       "      <td>6.813809</td>\n",
       "      <td>6.806788</td>\n",
       "      <td>6.837591</td>\n",
       "      <td>6.889389</td>\n",
       "      <td>6.837337</td>\n",
       "      <td>6.812384</td>\n",
       "      <td>6.746167</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47323 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM1902130  GSM1902131  GSM1902132  GSM1902133  GSM1902134  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   14.723105   14.974196   14.875049   14.262832   14.380508   \n",
       "ILMN_1343295   10.796875   10.545005   10.617579   10.663274   10.373557   \n",
       "ILMN_1651199    6.625879    6.658650    6.614950    6.729367    6.670856   \n",
       "ILMN_1651209    6.762562    6.775621    6.630290    6.822824    6.670856   \n",
       "ILMN_1651210    6.800612    6.596698    6.613691    6.504107    6.708310   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_3311170    6.598350    6.656326    6.745738    6.566161    6.572376   \n",
       "ILMN_3311175    6.652280    6.639231    6.695448    6.540806    6.721645   \n",
       "ILMN_3311180    6.864999    6.839016    6.642746    6.691449    6.716093   \n",
       "ILMN_3311185    6.632191    6.655011    6.718722    6.732387    6.545948   \n",
       "ILMN_3311190    6.709582    6.883397    6.685806    6.922629    6.997787   \n",
       "\n",
       "              GSM1902135  GSM1902136  GSM1902137  GSM1902138  GSM1902139  ...  \\\n",
       "ID                                                                        ...   \n",
       "ILMN_1343291   14.737535   14.994368   14.517866   14.751615   14.937985  ...   \n",
       "ILMN_1343295   11.070106   10.609179   10.614307   10.612894   10.263954  ...   \n",
       "ILMN_1651199    6.751337    6.658062    6.658457    6.644127    6.698253  ...   \n",
       "ILMN_1651209    6.713077    6.704106    6.854264    6.767462    6.659816  ...   \n",
       "ILMN_1651210    6.593606    6.674507    6.741474    6.683399    6.752126  ...   \n",
       "...                  ...         ...         ...         ...         ...  ...   \n",
       "ILMN_3311170    6.582429    6.613186    6.614191    6.634587    6.679710  ...   \n",
       "ILMN_3311175    6.662629    6.662562    6.680551    6.638602    6.637241  ...   \n",
       "ILMN_3311180    6.657982    6.662562    6.669022    6.671518    6.704380  ...   \n",
       "ILMN_3311185    6.675332    6.608312    6.635070    6.621255    6.705511  ...   \n",
       "ILMN_3311190    6.838690    6.824616    6.937539    6.860919    6.901888  ...   \n",
       "\n",
       "              GSM1902192  GSM1902193  GSM1902194  GSM1902195  GSM1902196  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   14.546172   14.609772   14.502806   14.751615   15.046889   \n",
       "ILMN_1343295   10.551885   10.677234   10.606567   10.688255   10.647923   \n",
       "ILMN_1651199    6.663772    6.603572    6.649300    6.763689    6.677779   \n",
       "ILMN_1651209    6.835929    6.729711    6.770642    6.811416    6.824817   \n",
       "ILMN_1651210    6.612110    6.662521    6.708362    6.758281    6.690449   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_3311170    6.665184    6.615608    6.665901    6.591221    6.563782   \n",
       "ILMN_3311175    6.673474    6.650218    6.685569    6.740448    6.637933   \n",
       "ILMN_3311180    6.770264    6.613962    6.695612    6.821868    6.648760   \n",
       "ILMN_3311185    6.742424    6.655760    6.705492    6.636093    6.659603   \n",
       "ILMN_3311190    6.978754    7.081290    7.009742    6.813809    6.806788   \n",
       "\n",
       "              GSM1902197  GSM1902198  GSM1902199  GSM1902200  GSM1902201  \n",
       "ID                                                                        \n",
       "ILMN_1343291   14.751615   14.890210   14.859273   14.937985   14.455832  \n",
       "ILMN_1343295   10.527452   10.519320   10.395747   10.680664   10.368892  \n",
       "ILMN_1651199    6.550622    6.693903    6.742800    6.584209    6.593124  \n",
       "ILMN_1651209    6.736938    6.766212    6.765388    6.690326    6.719627  \n",
       "ILMN_1651210    6.609686    6.743372    6.715278    6.601116    6.584874  \n",
       "...                  ...         ...         ...         ...         ...  \n",
       "ILMN_3311170    6.734294    6.627239    6.496727    6.703297    6.643118  \n",
       "ILMN_3311175    6.638491    6.666933    6.588203    6.743888    6.641553  \n",
       "ILMN_3311180    6.698223    6.615875    6.737687    6.725904    6.682509  \n",
       "ILMN_3311185    6.696920    6.622859    6.716543    6.626361    6.704373  \n",
       "ILMN_3311190    6.837591    6.889389    6.837337    6.812384    6.746167  \n",
       "\n",
       "[47323 rows x 72 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651199', 'ILMN_1651209', 'ILMN_1651210', 'ILMN_1651221', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651230', 'ILMN_1651232', 'ILMN_1651235', 'ILMN_1651236', 'ILMN_1651237', 'ILMN_1651238', 'ILMN_1651249', 'ILMN_1651253', 'ILMN_1651254', 'ILMN_1651259', 'ILMN_1651260', 'ILMN_1651262']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Source</th>\n",
       "      <th>Search_Key</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>ILMN_Gene</th>\n",
       "      <th>Source_Reference_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>Unigene_ID</th>\n",
       "      <th>Entrez_Gene_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Probe_Chr_Orientation</th>\n",
       "      <th>Probe_Coordinates</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Ontology_Component</th>\n",
       "      <th>Ontology_Process</th>\n",
       "      <th>Ontology_Function</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Obsolete_Probe_Id</th>\n",
       "      <th>GB_ACC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ILMN_1343048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ILMN_1343049</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ILMN_1343050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ILMN_1343052</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ILMN_1343059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3455430</th>\n",
       "      <td>ILMN_1908807</td>\n",
       "      <td>6.766003</td>\n",
       "      <td>0.17013</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3455431</th>\n",
       "      <td>ILMN_1701127</td>\n",
       "      <td>7.081023</td>\n",
       "      <td>0.0039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3455432</th>\n",
       "      <td>ILMN_1751164</td>\n",
       "      <td>7.821392</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3455433</th>\n",
       "      <td>ILMN_3296040</td>\n",
       "      <td>6.7245464</td>\n",
       "      <td>0.27532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3455434</th>\n",
       "      <td>ILMN_1693941</td>\n",
       "      <td>6.6382866</td>\n",
       "      <td>0.61948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3455435 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID    Species   Source Search_Key Transcript ILMN_Gene  \\\n",
       "0        ILMN_1343048        NaN      NaN        NaN        NaN       NaN   \n",
       "1        ILMN_1343049        NaN      NaN        NaN        NaN       NaN   \n",
       "2        ILMN_1343050        NaN      NaN        NaN        NaN       NaN   \n",
       "3        ILMN_1343052        NaN      NaN        NaN        NaN       NaN   \n",
       "4        ILMN_1343059        NaN      NaN        NaN        NaN       NaN   \n",
       "...               ...        ...      ...        ...        ...       ...   \n",
       "3455430  ILMN_1908807   6.766003  0.17013        NaN        NaN       NaN   \n",
       "3455431  ILMN_1701127   7.081023   0.0039        NaN        NaN       NaN   \n",
       "3455432  ILMN_1751164   7.821392        0        NaN        NaN       NaN   \n",
       "3455433  ILMN_3296040  6.7245464  0.27532        NaN        NaN       NaN   \n",
       "3455434  ILMN_1693941  6.6382866  0.61948        NaN        NaN       NaN   \n",
       "\n",
       "        Source_Reference_ID RefSeq_ID Unigene_ID  Entrez_Gene_ID  ...  \\\n",
       "0                       NaN       NaN        NaN             NaN  ...   \n",
       "1                       NaN       NaN        NaN             NaN  ...   \n",
       "2                       NaN       NaN        NaN             NaN  ...   \n",
       "3                       NaN       NaN        NaN             NaN  ...   \n",
       "4                       NaN       NaN        NaN             NaN  ...   \n",
       "...                     ...       ...        ...             ...  ...   \n",
       "3455430                 NaN       NaN        NaN             NaN  ...   \n",
       "3455431                 NaN       NaN        NaN             NaN  ...   \n",
       "3455432                 NaN       NaN        NaN             NaN  ...   \n",
       "3455433                 NaN       NaN        NaN             NaN  ...   \n",
       "3455434                 NaN       NaN        NaN             NaN  ...   \n",
       "\n",
       "         Probe_Chr_Orientation Probe_Coordinates Cytoband Definition  \\\n",
       "0                          NaN               NaN      NaN        NaN   \n",
       "1                          NaN               NaN      NaN        NaN   \n",
       "2                          NaN               NaN      NaN        NaN   \n",
       "3                          NaN               NaN      NaN        NaN   \n",
       "4                          NaN               NaN      NaN        NaN   \n",
       "...                        ...               ...      ...        ...   \n",
       "3455430                    NaN               NaN      NaN        NaN   \n",
       "3455431                    NaN               NaN      NaN        NaN   \n",
       "3455432                    NaN               NaN      NaN        NaN   \n",
       "3455433                    NaN               NaN      NaN        NaN   \n",
       "3455434                    NaN               NaN      NaN        NaN   \n",
       "\n",
       "        Ontology_Component  Ontology_Process Ontology_Function  Synonyms  \\\n",
       "0                      NaN               NaN               NaN       NaN   \n",
       "1                      NaN               NaN               NaN       NaN   \n",
       "2                      NaN               NaN               NaN       NaN   \n",
       "3                      NaN               NaN               NaN       NaN   \n",
       "4                      NaN               NaN               NaN       NaN   \n",
       "...                    ...               ...               ...       ...   \n",
       "3455430                NaN               NaN               NaN       NaN   \n",
       "3455431                NaN               NaN               NaN       NaN   \n",
       "3455432                NaN               NaN               NaN       NaN   \n",
       "3455433                NaN               NaN               NaN       NaN   \n",
       "3455434                NaN               NaN               NaN       NaN   \n",
       "\n",
       "        Obsolete_Probe_Id GB_ACC  \n",
       "0                     NaN    NaN  \n",
       "1                     NaN    NaN  \n",
       "2                     NaN    NaN  \n",
       "3                     NaN    NaN  \n",
       "4                     NaN    NaN  \n",
       "...                   ...    ...  \n",
       "3455430               NaN    NaN  \n",
       "3455431               NaN    NaN  \n",
       "3455432               NaN    NaN  \n",
       "3455433               NaN    NaN  \n",
       "3455434               NaN    NaN  \n",
       "\n",
       "[3455435 rows x 30 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Species', 'Source', 'Search_Key', 'Transcript', 'ILMN_Gene',\n",
       "       'Source_Reference_ID', 'RefSeq_ID', 'Unigene_ID', 'Entrez_Gene_ID',\n",
       "       'GI', 'Accession', 'Symbol', 'Protein_Product', 'Probe_Id',\n",
       "       'Array_Address_Id', 'Probe_Type', 'Probe_Start', 'SEQUENCE',\n",
       "       'Chromosome', 'Probe_Chr_Orientation', 'Probe_Coordinates', 'Cytoband',\n",
       "       'Definition', 'Ontology_Component', 'Ontology_Process',\n",
       "       'Ontology_Function', 'Synonyms', 'Obsolete_Probe_Id', 'GB_ACC'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1902130</th>\n",
       "      <th>GSM1902131</th>\n",
       "      <th>GSM1902132</th>\n",
       "      <th>GSM1902133</th>\n",
       "      <th>GSM1902134</th>\n",
       "      <th>GSM1902135</th>\n",
       "      <th>GSM1902136</th>\n",
       "      <th>GSM1902137</th>\n",
       "      <th>GSM1902138</th>\n",
       "      <th>GSM1902139</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1902192</th>\n",
       "      <th>GSM1902193</th>\n",
       "      <th>GSM1902194</th>\n",
       "      <th>GSM1902195</th>\n",
       "      <th>GSM1902196</th>\n",
       "      <th>GSM1902197</th>\n",
       "      <th>GSM1902198</th>\n",
       "      <th>GSM1902199</th>\n",
       "      <th>GSM1902200</th>\n",
       "      <th>GSM1902201</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1-Dec</th>\n",
       "      <td>6.583238</td>\n",
       "      <td>6.634187</td>\n",
       "      <td>6.736772</td>\n",
       "      <td>6.606144</td>\n",
       "      <td>6.592335</td>\n",
       "      <td>6.582429</td>\n",
       "      <td>6.727200</td>\n",
       "      <td>6.687715</td>\n",
       "      <td>6.815838</td>\n",
       "      <td>6.686119</td>\n",
       "      <td>...</td>\n",
       "      <td>6.730622</td>\n",
       "      <td>6.662521</td>\n",
       "      <td>6.536256</td>\n",
       "      <td>6.717677</td>\n",
       "      <td>6.567441</td>\n",
       "      <td>6.684505</td>\n",
       "      <td>6.635338</td>\n",
       "      <td>6.501363</td>\n",
       "      <td>6.668359</td>\n",
       "      <td>6.601340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-Mar</th>\n",
       "      <td>7.541307</td>\n",
       "      <td>7.418936</td>\n",
       "      <td>7.723857</td>\n",
       "      <td>7.457889</td>\n",
       "      <td>8.063496</td>\n",
       "      <td>7.345468</td>\n",
       "      <td>7.918340</td>\n",
       "      <td>7.683813</td>\n",
       "      <td>7.517664</td>\n",
       "      <td>7.331091</td>\n",
       "      <td>...</td>\n",
       "      <td>7.727845</td>\n",
       "      <td>7.882471</td>\n",
       "      <td>7.646876</td>\n",
       "      <td>7.620931</td>\n",
       "      <td>7.521968</td>\n",
       "      <td>7.612260</td>\n",
       "      <td>7.677948</td>\n",
       "      <td>7.704807</td>\n",
       "      <td>7.747943</td>\n",
       "      <td>7.453524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10-Mar</th>\n",
       "      <td>6.637913</td>\n",
       "      <td>6.584239</td>\n",
       "      <td>6.664167</td>\n",
       "      <td>6.665900</td>\n",
       "      <td>6.645813</td>\n",
       "      <td>6.645117</td>\n",
       "      <td>6.615416</td>\n",
       "      <td>6.647268</td>\n",
       "      <td>6.654260</td>\n",
       "      <td>6.575992</td>\n",
       "      <td>...</td>\n",
       "      <td>6.565555</td>\n",
       "      <td>6.667528</td>\n",
       "      <td>6.646483</td>\n",
       "      <td>6.655234</td>\n",
       "      <td>6.561255</td>\n",
       "      <td>6.596798</td>\n",
       "      <td>6.652807</td>\n",
       "      <td>6.647284</td>\n",
       "      <td>6.658114</td>\n",
       "      <td>6.607802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11-Mar</th>\n",
       "      <td>6.803986</td>\n",
       "      <td>6.617751</td>\n",
       "      <td>6.677545</td>\n",
       "      <td>6.577254</td>\n",
       "      <td>6.725436</td>\n",
       "      <td>6.696657</td>\n",
       "      <td>6.768100</td>\n",
       "      <td>6.588689</td>\n",
       "      <td>6.622732</td>\n",
       "      <td>6.529279</td>\n",
       "      <td>...</td>\n",
       "      <td>6.641082</td>\n",
       "      <td>6.693208</td>\n",
       "      <td>6.656019</td>\n",
       "      <td>6.701447</td>\n",
       "      <td>6.622160</td>\n",
       "      <td>6.628750</td>\n",
       "      <td>6.712999</td>\n",
       "      <td>6.660842</td>\n",
       "      <td>6.612525</td>\n",
       "      <td>6.849221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-Mar</th>\n",
       "      <td>7.515051</td>\n",
       "      <td>7.646111</td>\n",
       "      <td>7.551643</td>\n",
       "      <td>8.093898</td>\n",
       "      <td>7.642950</td>\n",
       "      <td>7.756180</td>\n",
       "      <td>7.660852</td>\n",
       "      <td>7.634819</td>\n",
       "      <td>8.171752</td>\n",
       "      <td>7.752842</td>\n",
       "      <td>...</td>\n",
       "      <td>7.408531</td>\n",
       "      <td>7.456631</td>\n",
       "      <td>7.849837</td>\n",
       "      <td>7.341353</td>\n",
       "      <td>7.283743</td>\n",
       "      <td>7.821662</td>\n",
       "      <td>7.364903</td>\n",
       "      <td>7.319059</td>\n",
       "      <td>7.650863</td>\n",
       "      <td>7.850871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dJ341D10.1</th>\n",
       "      <td>7.977261</td>\n",
       "      <td>8.056454</td>\n",
       "      <td>7.997424</td>\n",
       "      <td>8.412908</td>\n",
       "      <td>8.033030</td>\n",
       "      <td>7.717434</td>\n",
       "      <td>7.861945</td>\n",
       "      <td>8.140966</td>\n",
       "      <td>7.782730</td>\n",
       "      <td>7.924043</td>\n",
       "      <td>...</td>\n",
       "      <td>7.971788</td>\n",
       "      <td>8.230297</td>\n",
       "      <td>7.978352</td>\n",
       "      <td>7.938903</td>\n",
       "      <td>7.934387</td>\n",
       "      <td>7.668821</td>\n",
       "      <td>7.950748</td>\n",
       "      <td>7.760117</td>\n",
       "      <td>7.918305</td>\n",
       "      <td>7.816753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gm127</th>\n",
       "      <td>6.849804</td>\n",
       "      <td>6.806737</td>\n",
       "      <td>6.851868</td>\n",
       "      <td>6.970032</td>\n",
       "      <td>6.857639</td>\n",
       "      <td>6.631670</td>\n",
       "      <td>6.785138</td>\n",
       "      <td>6.753465</td>\n",
       "      <td>6.807225</td>\n",
       "      <td>6.670449</td>\n",
       "      <td>...</td>\n",
       "      <td>6.737225</td>\n",
       "      <td>6.884189</td>\n",
       "      <td>6.812922</td>\n",
       "      <td>6.720719</td>\n",
       "      <td>6.863619</td>\n",
       "      <td>6.722490</td>\n",
       "      <td>6.738746</td>\n",
       "      <td>6.873762</td>\n",
       "      <td>6.905868</td>\n",
       "      <td>6.845744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psiTPTE22</th>\n",
       "      <td>6.817363</td>\n",
       "      <td>6.632995</td>\n",
       "      <td>6.764498</td>\n",
       "      <td>6.673868</td>\n",
       "      <td>6.815591</td>\n",
       "      <td>6.803827</td>\n",
       "      <td>6.751432</td>\n",
       "      <td>6.836163</td>\n",
       "      <td>6.743703</td>\n",
       "      <td>6.660348</td>\n",
       "      <td>...</td>\n",
       "      <td>6.831514</td>\n",
       "      <td>6.734106</td>\n",
       "      <td>6.700657</td>\n",
       "      <td>6.713387</td>\n",
       "      <td>6.765148</td>\n",
       "      <td>6.750495</td>\n",
       "      <td>6.787045</td>\n",
       "      <td>6.755791</td>\n",
       "      <td>6.733667</td>\n",
       "      <td>6.791706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rab1c</th>\n",
       "      <td>6.741328</td>\n",
       "      <td>6.624208</td>\n",
       "      <td>6.761005</td>\n",
       "      <td>6.754579</td>\n",
       "      <td>6.889446</td>\n",
       "      <td>6.790038</td>\n",
       "      <td>6.850925</td>\n",
       "      <td>6.762543</td>\n",
       "      <td>6.914238</td>\n",
       "      <td>6.925648</td>\n",
       "      <td>...</td>\n",
       "      <td>6.799939</td>\n",
       "      <td>6.769505</td>\n",
       "      <td>6.811423</td>\n",
       "      <td>6.713453</td>\n",
       "      <td>6.852134</td>\n",
       "      <td>6.777274</td>\n",
       "      <td>6.651200</td>\n",
       "      <td>6.628250</td>\n",
       "      <td>6.637634</td>\n",
       "      <td>6.675023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tAKR</th>\n",
       "      <td>6.649673</td>\n",
       "      <td>6.651072</td>\n",
       "      <td>6.737572</td>\n",
       "      <td>6.619973</td>\n",
       "      <td>6.628795</td>\n",
       "      <td>6.668012</td>\n",
       "      <td>6.642919</td>\n",
       "      <td>6.709250</td>\n",
       "      <td>6.694066</td>\n",
       "      <td>6.616257</td>\n",
       "      <td>...</td>\n",
       "      <td>6.658659</td>\n",
       "      <td>6.699603</td>\n",
       "      <td>6.634119</td>\n",
       "      <td>6.679162</td>\n",
       "      <td>6.638257</td>\n",
       "      <td>6.711074</td>\n",
       "      <td>6.659563</td>\n",
       "      <td>6.741700</td>\n",
       "      <td>6.594904</td>\n",
       "      <td>6.639587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31426 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM1902130  GSM1902131  GSM1902132  GSM1902133  GSM1902134  \\\n",
       "Gene                                                                     \n",
       "1-Dec         6.583238    6.634187    6.736772    6.606144    6.592335   \n",
       "1-Mar         7.541307    7.418936    7.723857    7.457889    8.063496   \n",
       "10-Mar        6.637913    6.584239    6.664167    6.665900    6.645813   \n",
       "11-Mar        6.803986    6.617751    6.677545    6.577254    6.725436   \n",
       "2-Mar         7.515051    7.646111    7.551643    8.093898    7.642950   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "dJ341D10.1    7.977261    8.056454    7.997424    8.412908    8.033030   \n",
       "gm127         6.849804    6.806737    6.851868    6.970032    6.857639   \n",
       "psiTPTE22     6.817363    6.632995    6.764498    6.673868    6.815591   \n",
       "rab1c         6.741328    6.624208    6.761005    6.754579    6.889446   \n",
       "tAKR          6.649673    6.651072    6.737572    6.619973    6.628795   \n",
       "\n",
       "            GSM1902135  GSM1902136  GSM1902137  GSM1902138  GSM1902139  ...  \\\n",
       "Gene                                                                    ...   \n",
       "1-Dec         6.582429    6.727200    6.687715    6.815838    6.686119  ...   \n",
       "1-Mar         7.345468    7.918340    7.683813    7.517664    7.331091  ...   \n",
       "10-Mar        6.645117    6.615416    6.647268    6.654260    6.575992  ...   \n",
       "11-Mar        6.696657    6.768100    6.588689    6.622732    6.529279  ...   \n",
       "2-Mar         7.756180    7.660852    7.634819    8.171752    7.752842  ...   \n",
       "...                ...         ...         ...         ...         ...  ...   \n",
       "dJ341D10.1    7.717434    7.861945    8.140966    7.782730    7.924043  ...   \n",
       "gm127         6.631670    6.785138    6.753465    6.807225    6.670449  ...   \n",
       "psiTPTE22     6.803827    6.751432    6.836163    6.743703    6.660348  ...   \n",
       "rab1c         6.790038    6.850925    6.762543    6.914238    6.925648  ...   \n",
       "tAKR          6.668012    6.642919    6.709250    6.694066    6.616257  ...   \n",
       "\n",
       "            GSM1902192  GSM1902193  GSM1902194  GSM1902195  GSM1902196  \\\n",
       "Gene                                                                     \n",
       "1-Dec         6.730622    6.662521    6.536256    6.717677    6.567441   \n",
       "1-Mar         7.727845    7.882471    7.646876    7.620931    7.521968   \n",
       "10-Mar        6.565555    6.667528    6.646483    6.655234    6.561255   \n",
       "11-Mar        6.641082    6.693208    6.656019    6.701447    6.622160   \n",
       "2-Mar         7.408531    7.456631    7.849837    7.341353    7.283743   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "dJ341D10.1    7.971788    8.230297    7.978352    7.938903    7.934387   \n",
       "gm127         6.737225    6.884189    6.812922    6.720719    6.863619   \n",
       "psiTPTE22     6.831514    6.734106    6.700657    6.713387    6.765148   \n",
       "rab1c         6.799939    6.769505    6.811423    6.713453    6.852134   \n",
       "tAKR          6.658659    6.699603    6.634119    6.679162    6.638257   \n",
       "\n",
       "            GSM1902197  GSM1902198  GSM1902199  GSM1902200  GSM1902201  \n",
       "Gene                                                                    \n",
       "1-Dec         6.684505    6.635338    6.501363    6.668359    6.601340  \n",
       "1-Mar         7.612260    7.677948    7.704807    7.747943    7.453524  \n",
       "10-Mar        6.596798    6.652807    6.647284    6.658114    6.607802  \n",
       "11-Mar        6.628750    6.712999    6.660842    6.612525    6.849221  \n",
       "2-Mar         7.821662    7.364903    7.319059    7.650863    7.850871  \n",
       "...                ...         ...         ...         ...         ...  \n",
       "dJ341D10.1    7.668821    7.950748    7.760117    7.918305    7.816753  \n",
       "gm127         6.722490    6.738746    6.873762    6.905868    6.845744  \n",
       "psiTPTE22     6.750495    6.787045    6.755791    6.733667    6.791706  \n",
       "rab1c         6.777274    6.651200    6.628250    6.637634    6.675023  \n",
       "tAKR          6.711074    6.659563    6.741700    6.594904    6.639587  \n",
       "\n",
       "[31426 rows x 72 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1902130</th>\n",
       "      <th>GSM1902131</th>\n",
       "      <th>GSM1902132</th>\n",
       "      <th>GSM1902133</th>\n",
       "      <th>GSM1902134</th>\n",
       "      <th>GSM1902135</th>\n",
       "      <th>GSM1902136</th>\n",
       "      <th>GSM1902137</th>\n",
       "      <th>GSM1902138</th>\n",
       "      <th>GSM1902139</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1902192</th>\n",
       "      <th>GSM1902193</th>\n",
       "      <th>GSM1902194</th>\n",
       "      <th>GSM1902195</th>\n",
       "      <th>GSM1902196</th>\n",
       "      <th>GSM1902197</th>\n",
       "      <th>GSM1902198</th>\n",
       "      <th>GSM1902199</th>\n",
       "      <th>GSM1902200</th>\n",
       "      <th>GSM1902201</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>6.721839</td>\n",
       "      <td>6.673086</td>\n",
       "      <td>6.712955</td>\n",
       "      <td>6.770714</td>\n",
       "      <td>6.712682</td>\n",
       "      <td>6.767250</td>\n",
       "      <td>6.680871</td>\n",
       "      <td>6.732433</td>\n",
       "      <td>6.645191</td>\n",
       "      <td>6.721483</td>\n",
       "      <td>...</td>\n",
       "      <td>6.663826</td>\n",
       "      <td>6.726559</td>\n",
       "      <td>6.688819</td>\n",
       "      <td>6.754805</td>\n",
       "      <td>6.705244</td>\n",
       "      <td>6.714713</td>\n",
       "      <td>6.752249</td>\n",
       "      <td>6.711358</td>\n",
       "      <td>6.759263</td>\n",
       "      <td>6.767227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>6.687158</td>\n",
       "      <td>6.624651</td>\n",
       "      <td>6.642954</td>\n",
       "      <td>6.664230</td>\n",
       "      <td>6.665765</td>\n",
       "      <td>6.677747</td>\n",
       "      <td>6.677909</td>\n",
       "      <td>6.644698</td>\n",
       "      <td>6.646873</td>\n",
       "      <td>6.668884</td>\n",
       "      <td>...</td>\n",
       "      <td>6.658412</td>\n",
       "      <td>6.649672</td>\n",
       "      <td>6.705482</td>\n",
       "      <td>6.721324</td>\n",
       "      <td>6.799274</td>\n",
       "      <td>6.621188</td>\n",
       "      <td>6.664272</td>\n",
       "      <td>6.739637</td>\n",
       "      <td>6.641758</td>\n",
       "      <td>6.708410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>6.619832</td>\n",
       "      <td>6.542803</td>\n",
       "      <td>6.522699</td>\n",
       "      <td>6.633948</td>\n",
       "      <td>6.562571</td>\n",
       "      <td>6.568287</td>\n",
       "      <td>6.528718</td>\n",
       "      <td>6.529343</td>\n",
       "      <td>6.533412</td>\n",
       "      <td>6.462207</td>\n",
       "      <td>...</td>\n",
       "      <td>6.565053</td>\n",
       "      <td>6.575708</td>\n",
       "      <td>6.484352</td>\n",
       "      <td>6.526597</td>\n",
       "      <td>6.537614</td>\n",
       "      <td>6.520083</td>\n",
       "      <td>6.586989</td>\n",
       "      <td>6.711289</td>\n",
       "      <td>6.548265</td>\n",
       "      <td>6.591539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>6.567780</td>\n",
       "      <td>6.488923</td>\n",
       "      <td>6.513175</td>\n",
       "      <td>6.599086</td>\n",
       "      <td>6.622063</td>\n",
       "      <td>6.612026</td>\n",
       "      <td>6.483853</td>\n",
       "      <td>6.690247</td>\n",
       "      <td>6.527108</td>\n",
       "      <td>6.492720</td>\n",
       "      <td>...</td>\n",
       "      <td>6.509235</td>\n",
       "      <td>6.595049</td>\n",
       "      <td>6.582683</td>\n",
       "      <td>6.524981</td>\n",
       "      <td>6.558844</td>\n",
       "      <td>6.676249</td>\n",
       "      <td>6.471403</td>\n",
       "      <td>6.618289</td>\n",
       "      <td>6.416780</td>\n",
       "      <td>6.467222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>6.614632</td>\n",
       "      <td>6.741219</td>\n",
       "      <td>6.615549</td>\n",
       "      <td>6.703858</td>\n",
       "      <td>6.693452</td>\n",
       "      <td>6.785090</td>\n",
       "      <td>6.611774</td>\n",
       "      <td>6.734135</td>\n",
       "      <td>6.758213</td>\n",
       "      <td>6.595013</td>\n",
       "      <td>...</td>\n",
       "      <td>6.613405</td>\n",
       "      <td>6.615288</td>\n",
       "      <td>6.633757</td>\n",
       "      <td>6.611420</td>\n",
       "      <td>6.658956</td>\n",
       "      <td>6.751886</td>\n",
       "      <td>6.724360</td>\n",
       "      <td>6.759662</td>\n",
       "      <td>6.663325</td>\n",
       "      <td>6.663378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>8.893455</td>\n",
       "      <td>9.275059</td>\n",
       "      <td>9.146032</td>\n",
       "      <td>8.504040</td>\n",
       "      <td>8.506646</td>\n",
       "      <td>9.501809</td>\n",
       "      <td>8.781733</td>\n",
       "      <td>8.942747</td>\n",
       "      <td>9.295257</td>\n",
       "      <td>8.939878</td>\n",
       "      <td>...</td>\n",
       "      <td>9.292432</td>\n",
       "      <td>8.983416</td>\n",
       "      <td>9.144604</td>\n",
       "      <td>8.913773</td>\n",
       "      <td>8.860433</td>\n",
       "      <td>8.674577</td>\n",
       "      <td>8.972839</td>\n",
       "      <td>8.776298</td>\n",
       "      <td>8.944420</td>\n",
       "      <td>8.943609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>10.224247</td>\n",
       "      <td>10.687603</td>\n",
       "      <td>10.355187</td>\n",
       "      <td>10.090020</td>\n",
       "      <td>10.490643</td>\n",
       "      <td>10.181591</td>\n",
       "      <td>10.305449</td>\n",
       "      <td>10.251164</td>\n",
       "      <td>10.484419</td>\n",
       "      <td>10.033065</td>\n",
       "      <td>...</td>\n",
       "      <td>9.807033</td>\n",
       "      <td>10.682264</td>\n",
       "      <td>10.716734</td>\n",
       "      <td>10.330274</td>\n",
       "      <td>10.143487</td>\n",
       "      <td>10.200066</td>\n",
       "      <td>10.003246</td>\n",
       "      <td>9.996454</td>\n",
       "      <td>10.237811</td>\n",
       "      <td>10.164364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>8.439836</td>\n",
       "      <td>8.507540</td>\n",
       "      <td>8.215377</td>\n",
       "      <td>8.177778</td>\n",
       "      <td>8.198018</td>\n",
       "      <td>8.288051</td>\n",
       "      <td>7.965330</td>\n",
       "      <td>8.336710</td>\n",
       "      <td>8.172971</td>\n",
       "      <td>8.184372</td>\n",
       "      <td>...</td>\n",
       "      <td>8.381381</td>\n",
       "      <td>8.294295</td>\n",
       "      <td>8.237680</td>\n",
       "      <td>8.304722</td>\n",
       "      <td>8.183370</td>\n",
       "      <td>8.010283</td>\n",
       "      <td>8.331811</td>\n",
       "      <td>8.261510</td>\n",
       "      <td>8.301020</td>\n",
       "      <td>8.169597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>7.975543</td>\n",
       "      <td>7.979056</td>\n",
       "      <td>7.773221</td>\n",
       "      <td>7.542429</td>\n",
       "      <td>7.725909</td>\n",
       "      <td>7.796597</td>\n",
       "      <td>7.886633</td>\n",
       "      <td>7.822113</td>\n",
       "      <td>7.877351</td>\n",
       "      <td>7.945854</td>\n",
       "      <td>...</td>\n",
       "      <td>7.869515</td>\n",
       "      <td>7.740685</td>\n",
       "      <td>7.638251</td>\n",
       "      <td>8.012779</td>\n",
       "      <td>7.999851</td>\n",
       "      <td>7.920978</td>\n",
       "      <td>7.926523</td>\n",
       "      <td>8.024448</td>\n",
       "      <td>7.776912</td>\n",
       "      <td>7.721161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB1C</th>\n",
       "      <td>6.741328</td>\n",
       "      <td>6.624208</td>\n",
       "      <td>6.761005</td>\n",
       "      <td>6.754579</td>\n",
       "      <td>6.889446</td>\n",
       "      <td>6.790038</td>\n",
       "      <td>6.850925</td>\n",
       "      <td>6.762543</td>\n",
       "      <td>6.914238</td>\n",
       "      <td>6.925648</td>\n",
       "      <td>...</td>\n",
       "      <td>6.799939</td>\n",
       "      <td>6.769505</td>\n",
       "      <td>6.811423</td>\n",
       "      <td>6.713453</td>\n",
       "      <td>6.852134</td>\n",
       "      <td>6.777274</td>\n",
       "      <td>6.651200</td>\n",
       "      <td>6.628250</td>\n",
       "      <td>6.637634</td>\n",
       "      <td>6.675023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17891 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1902130  GSM1902131  GSM1902132  GSM1902133  GSM1902134  \\\n",
       "A1BG       6.721839    6.673086    6.712955    6.770714    6.712682   \n",
       "A1CF       6.687158    6.624651    6.642954    6.664230    6.665765   \n",
       "A2M        6.619832    6.542803    6.522699    6.633948    6.562571   \n",
       "A2ML1      6.567780    6.488923    6.513175    6.599086    6.622063   \n",
       "A3GALT2    6.614632    6.741219    6.615549    6.703858    6.693452   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B     8.893455    9.275059    9.146032    8.504040    8.506646   \n",
       "ZYX       10.224247   10.687603   10.355187   10.090020   10.490643   \n",
       "ZZEF1      8.439836    8.507540    8.215377    8.177778    8.198018   \n",
       "ZZZ3       7.975543    7.979056    7.773221    7.542429    7.725909   \n",
       "RAB1C      6.741328    6.624208    6.761005    6.754579    6.889446   \n",
       "\n",
       "         GSM1902135  GSM1902136  GSM1902137  GSM1902138  GSM1902139  ...  \\\n",
       "A1BG       6.767250    6.680871    6.732433    6.645191    6.721483  ...   \n",
       "A1CF       6.677747    6.677909    6.644698    6.646873    6.668884  ...   \n",
       "A2M        6.568287    6.528718    6.529343    6.533412    6.462207  ...   \n",
       "A2ML1      6.612026    6.483853    6.690247    6.527108    6.492720  ...   \n",
       "A3GALT2    6.785090    6.611774    6.734135    6.758213    6.595013  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZYG11B     9.501809    8.781733    8.942747    9.295257    8.939878  ...   \n",
       "ZYX       10.181591   10.305449   10.251164   10.484419   10.033065  ...   \n",
       "ZZEF1      8.288051    7.965330    8.336710    8.172971    8.184372  ...   \n",
       "ZZZ3       7.796597    7.886633    7.822113    7.877351    7.945854  ...   \n",
       "RAB1C      6.790038    6.850925    6.762543    6.914238    6.925648  ...   \n",
       "\n",
       "         GSM1902192  GSM1902193  GSM1902194  GSM1902195  GSM1902196  \\\n",
       "A1BG       6.663826    6.726559    6.688819    6.754805    6.705244   \n",
       "A1CF       6.658412    6.649672    6.705482    6.721324    6.799274   \n",
       "A2M        6.565053    6.575708    6.484352    6.526597    6.537614   \n",
       "A2ML1      6.509235    6.595049    6.582683    6.524981    6.558844   \n",
       "A3GALT2    6.613405    6.615288    6.633757    6.611420    6.658956   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B     9.292432    8.983416    9.144604    8.913773    8.860433   \n",
       "ZYX        9.807033   10.682264   10.716734   10.330274   10.143487   \n",
       "ZZEF1      8.381381    8.294295    8.237680    8.304722    8.183370   \n",
       "ZZZ3       7.869515    7.740685    7.638251    8.012779    7.999851   \n",
       "RAB1C      6.799939    6.769505    6.811423    6.713453    6.852134   \n",
       "\n",
       "         GSM1902197  GSM1902198  GSM1902199  GSM1902200  GSM1902201  \n",
       "A1BG       6.714713    6.752249    6.711358    6.759263    6.767227  \n",
       "A1CF       6.621188    6.664272    6.739637    6.641758    6.708410  \n",
       "A2M        6.520083    6.586989    6.711289    6.548265    6.591539  \n",
       "A2ML1      6.676249    6.471403    6.618289    6.416780    6.467222  \n",
       "A3GALT2    6.751886    6.724360    6.759662    6.663325    6.663378  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZYG11B     8.674577    8.972839    8.776298    8.944420    8.943609  \n",
       "ZYX       10.200066   10.003246    9.996454   10.237811   10.164364  \n",
       "ZZEF1      8.010283    8.331811    8.261510    8.301020    8.169597  \n",
       "ZZZ3       7.920978    7.926523    8.024448    7.776912    7.721161  \n",
       "RAB1C      6.777274    6.651200    6.628250    6.637634    6.675023  \n",
       "\n",
       "[17891 rows x 72 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>RAB1C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1902130</th>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.721839</td>\n",
       "      <td>6.687158</td>\n",
       "      <td>6.619832</td>\n",
       "      <td>6.567780</td>\n",
       "      <td>6.614632</td>\n",
       "      <td>6.689058</td>\n",
       "      <td>6.681622</td>\n",
       "      <td>...</td>\n",
       "      <td>6.705402</td>\n",
       "      <td>6.664771</td>\n",
       "      <td>6.928788</td>\n",
       "      <td>6.942867</td>\n",
       "      <td>6.616599</td>\n",
       "      <td>8.893455</td>\n",
       "      <td>10.224247</td>\n",
       "      <td>8.439836</td>\n",
       "      <td>7.975543</td>\n",
       "      <td>6.741328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902131</th>\n",
       "      <td>1.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.673086</td>\n",
       "      <td>6.624651</td>\n",
       "      <td>6.542803</td>\n",
       "      <td>6.488923</td>\n",
       "      <td>6.741219</td>\n",
       "      <td>6.659898</td>\n",
       "      <td>6.792750</td>\n",
       "      <td>...</td>\n",
       "      <td>6.683924</td>\n",
       "      <td>6.626405</td>\n",
       "      <td>6.902658</td>\n",
       "      <td>6.945891</td>\n",
       "      <td>6.701322</td>\n",
       "      <td>9.275059</td>\n",
       "      <td>10.687603</td>\n",
       "      <td>8.507540</td>\n",
       "      <td>7.979056</td>\n",
       "      <td>6.624208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902132</th>\n",
       "      <td>1.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.712955</td>\n",
       "      <td>6.642954</td>\n",
       "      <td>6.522699</td>\n",
       "      <td>6.513175</td>\n",
       "      <td>6.615549</td>\n",
       "      <td>6.638833</td>\n",
       "      <td>6.765411</td>\n",
       "      <td>...</td>\n",
       "      <td>6.668817</td>\n",
       "      <td>6.638077</td>\n",
       "      <td>6.995390</td>\n",
       "      <td>6.998783</td>\n",
       "      <td>6.642820</td>\n",
       "      <td>9.146032</td>\n",
       "      <td>10.355187</td>\n",
       "      <td>8.215377</td>\n",
       "      <td>7.773221</td>\n",
       "      <td>6.761005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902133</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.770714</td>\n",
       "      <td>6.664230</td>\n",
       "      <td>6.633948</td>\n",
       "      <td>6.599086</td>\n",
       "      <td>6.703858</td>\n",
       "      <td>6.620400</td>\n",
       "      <td>7.051553</td>\n",
       "      <td>...</td>\n",
       "      <td>6.675158</td>\n",
       "      <td>6.668629</td>\n",
       "      <td>6.711156</td>\n",
       "      <td>7.044865</td>\n",
       "      <td>6.705459</td>\n",
       "      <td>8.504040</td>\n",
       "      <td>10.090020</td>\n",
       "      <td>8.177778</td>\n",
       "      <td>7.542429</td>\n",
       "      <td>6.754579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902134</th>\n",
       "      <td>1.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.712682</td>\n",
       "      <td>6.665765</td>\n",
       "      <td>6.562571</td>\n",
       "      <td>6.622063</td>\n",
       "      <td>6.693452</td>\n",
       "      <td>6.586829</td>\n",
       "      <td>6.802776</td>\n",
       "      <td>...</td>\n",
       "      <td>6.736839</td>\n",
       "      <td>6.653528</td>\n",
       "      <td>6.818604</td>\n",
       "      <td>6.877028</td>\n",
       "      <td>6.696736</td>\n",
       "      <td>8.506646</td>\n",
       "      <td>10.490643</td>\n",
       "      <td>8.198018</td>\n",
       "      <td>7.725909</td>\n",
       "      <td>6.889446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902197</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.714713</td>\n",
       "      <td>6.621188</td>\n",
       "      <td>6.520083</td>\n",
       "      <td>6.676249</td>\n",
       "      <td>6.751886</td>\n",
       "      <td>6.580731</td>\n",
       "      <td>6.945715</td>\n",
       "      <td>...</td>\n",
       "      <td>6.731256</td>\n",
       "      <td>6.639304</td>\n",
       "      <td>6.856193</td>\n",
       "      <td>6.968547</td>\n",
       "      <td>6.762333</td>\n",
       "      <td>8.674577</td>\n",
       "      <td>10.200066</td>\n",
       "      <td>8.010283</td>\n",
       "      <td>7.920978</td>\n",
       "      <td>6.777274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902198</th>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.752249</td>\n",
       "      <td>6.664272</td>\n",
       "      <td>6.586989</td>\n",
       "      <td>6.471403</td>\n",
       "      <td>6.724360</td>\n",
       "      <td>6.556073</td>\n",
       "      <td>6.747668</td>\n",
       "      <td>...</td>\n",
       "      <td>6.723123</td>\n",
       "      <td>6.694954</td>\n",
       "      <td>6.942069</td>\n",
       "      <td>6.970328</td>\n",
       "      <td>6.651852</td>\n",
       "      <td>8.972839</td>\n",
       "      <td>10.003246</td>\n",
       "      <td>8.331811</td>\n",
       "      <td>7.926523</td>\n",
       "      <td>6.651200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902199</th>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.711358</td>\n",
       "      <td>6.739637</td>\n",
       "      <td>6.711289</td>\n",
       "      <td>6.618289</td>\n",
       "      <td>6.759662</td>\n",
       "      <td>6.596578</td>\n",
       "      <td>6.671845</td>\n",
       "      <td>...</td>\n",
       "      <td>6.684863</td>\n",
       "      <td>6.703465</td>\n",
       "      <td>6.790693</td>\n",
       "      <td>6.942059</td>\n",
       "      <td>6.648330</td>\n",
       "      <td>8.776298</td>\n",
       "      <td>9.996454</td>\n",
       "      <td>8.261510</td>\n",
       "      <td>8.024448</td>\n",
       "      <td>6.628250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902200</th>\n",
       "      <td>0.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.759263</td>\n",
       "      <td>6.641758</td>\n",
       "      <td>6.548265</td>\n",
       "      <td>6.416780</td>\n",
       "      <td>6.663325</td>\n",
       "      <td>6.565605</td>\n",
       "      <td>6.846892</td>\n",
       "      <td>...</td>\n",
       "      <td>6.649319</td>\n",
       "      <td>6.658996</td>\n",
       "      <td>6.859029</td>\n",
       "      <td>7.010749</td>\n",
       "      <td>6.641221</td>\n",
       "      <td>8.944420</td>\n",
       "      <td>10.237811</td>\n",
       "      <td>8.301020</td>\n",
       "      <td>7.776912</td>\n",
       "      <td>6.637634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1902201</th>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.767227</td>\n",
       "      <td>6.708410</td>\n",
       "      <td>6.591539</td>\n",
       "      <td>6.467222</td>\n",
       "      <td>6.663378</td>\n",
       "      <td>6.569665</td>\n",
       "      <td>6.716382</td>\n",
       "      <td>...</td>\n",
       "      <td>6.685942</td>\n",
       "      <td>6.617394</td>\n",
       "      <td>6.784757</td>\n",
       "      <td>7.064588</td>\n",
       "      <td>6.700598</td>\n",
       "      <td>8.943609</td>\n",
       "      <td>10.164364</td>\n",
       "      <td>8.169597</td>\n",
       "      <td>7.721161</td>\n",
       "      <td>6.675023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows Ã— 17894 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Ankylosing Spondylitis   Age  Gender      A1BG      A1CF  \\\n",
       "GSM1902130                     1.0  53.0     0.0  6.721839  6.687158   \n",
       "GSM1902131                     1.0  26.0     0.0  6.673086  6.624651   \n",
       "GSM1902132                     1.0  29.0     0.0  6.712955  6.642954   \n",
       "GSM1902133                     1.0  50.0     0.0  6.770714  6.664230   \n",
       "GSM1902134                     1.0  35.0     0.0  6.712682  6.665765   \n",
       "...                            ...   ...     ...       ...       ...   \n",
       "GSM1902197                     0.0  20.0     1.0  6.714713  6.621188   \n",
       "GSM1902198                     0.0  31.0     1.0  6.752249  6.664272   \n",
       "GSM1902199                     0.0  25.0     1.0  6.711358  6.739637   \n",
       "GSM1902200                     0.0  29.0     1.0  6.759263  6.641758   \n",
       "GSM1902201                     0.0  65.0     1.0  6.767227  6.708410   \n",
       "\n",
       "                 A2M     A2ML1   A3GALT2    A4GALT     A4GNT  ...     ZWINT  \\\n",
       "GSM1902130  6.619832  6.567780  6.614632  6.689058  6.681622  ...  6.705402   \n",
       "GSM1902131  6.542803  6.488923  6.741219  6.659898  6.792750  ...  6.683924   \n",
       "GSM1902132  6.522699  6.513175  6.615549  6.638833  6.765411  ...  6.668817   \n",
       "GSM1902133  6.633948  6.599086  6.703858  6.620400  7.051553  ...  6.675158   \n",
       "GSM1902134  6.562571  6.622063  6.693452  6.586829  6.802776  ...  6.736839   \n",
       "...              ...       ...       ...       ...       ...  ...       ...   \n",
       "GSM1902197  6.520083  6.676249  6.751886  6.580731  6.945715  ...  6.731256   \n",
       "GSM1902198  6.586989  6.471403  6.724360  6.556073  6.747668  ...  6.723123   \n",
       "GSM1902199  6.711289  6.618289  6.759662  6.596578  6.671845  ...  6.684863   \n",
       "GSM1902200  6.548265  6.416780  6.663325  6.565605  6.846892  ...  6.649319   \n",
       "GSM1902201  6.591539  6.467222  6.663378  6.569665  6.716382  ...  6.685942   \n",
       "\n",
       "                ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B        ZYX  \\\n",
       "GSM1902130  6.664771  6.928788  6.942867  6.616599  8.893455  10.224247   \n",
       "GSM1902131  6.626405  6.902658  6.945891  6.701322  9.275059  10.687603   \n",
       "GSM1902132  6.638077  6.995390  6.998783  6.642820  9.146032  10.355187   \n",
       "GSM1902133  6.668629  6.711156  7.044865  6.705459  8.504040  10.090020   \n",
       "GSM1902134  6.653528  6.818604  6.877028  6.696736  8.506646  10.490643   \n",
       "...              ...       ...       ...       ...       ...        ...   \n",
       "GSM1902197  6.639304  6.856193  6.968547  6.762333  8.674577  10.200066   \n",
       "GSM1902198  6.694954  6.942069  6.970328  6.651852  8.972839  10.003246   \n",
       "GSM1902199  6.703465  6.790693  6.942059  6.648330  8.776298   9.996454   \n",
       "GSM1902200  6.658996  6.859029  7.010749  6.641221  8.944420  10.237811   \n",
       "GSM1902201  6.617394  6.784757  7.064588  6.700598  8.943609  10.164364   \n",
       "\n",
       "               ZZEF1      ZZZ3     RAB1C  \n",
       "GSM1902130  8.439836  7.975543  6.741328  \n",
       "GSM1902131  8.507540  7.979056  6.624208  \n",
       "GSM1902132  8.215377  7.773221  6.761005  \n",
       "GSM1902133  8.177778  7.542429  6.754579  \n",
       "GSM1902134  8.198018  7.725909  6.889446  \n",
       "...              ...       ...       ...  \n",
       "GSM1902197  8.010283  7.920978  6.777274  \n",
       "GSM1902198  8.331811  7.926523  6.651200  \n",
       "GSM1902199  8.261510  8.024448  6.628250  \n",
       "GSM1902200  8.301020  7.776912  6.637634  \n",
       "GSM1902201  8.169597  7.721161  6.675023  \n",
       "\n",
       "[72 rows x 17894 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 72 samples.\n",
      "For the feature 'Ankylosing Spondylitis', the least common label is '0.0' with 20 occurrences. This represents 27.78% of the dataset.\n",
      "The distribution of the feature 'Ankylosing Spondylitis' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 28.75\n",
      "  50% (Median): 41.5\n",
      "  75%: 51.25\n",
      "Min: 18.0\n",
      "Max: 77.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 35 occurrences. This represents 48.61% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression analysis of macrophages from ankylosing spondylitis patients reveals interferon-gamma dysregulation\"\n",
      "!Series_summary\t\"OBJECTIVE: To determine whether macrophages, a type of cell implicated in the pathogenesis of ankylosing spondylitis (AS), exhibit a characteristic gene expression pattern. METHODS: Macrophages were derived from the peripheral blood of 8 AS patients (median disease duration 13 years [range <1-43 years]) and 9 healthy control subjects over 7 days with the use of granulocyte-macrophage colony-stimulating factor. Cells were stimulated for 24 hours with interferon-gamma (IFNgamma; 100 units/ml), were left untreated for 24 hours, or were treated for 3 hours with lipopolysaccharide (LPS; 10 ng/ml). RNA was isolated and examined by microarray and real-time quantitative reverse transcription-polymerase chain reaction analysis. RESULTS: Microarray analysis revealed 198 probe sets detecting the differential expression of 141 unique genes in untreated macrophages from AS patients compared with healthy controls. Clustering and principal components analysis clearly distinguished AS patients and controls. Of the differentially expressed genes, 78 (55%) were IFN-regulated, and their relative expression indicated a reverse IFN signature in AS patient macrophages, where IFNgamma-up-regulated genes were underexpressed and down-regulated genes were overexpressed. Treatment of macrophages with exogenous IFNgamma normalized the expression of these genes between patients and controls. In addition, the messenger RNA encoded by the IFNgamma gene was approximately 2-fold lower in AS patient macrophages at baseline (P = 0.004) and was poorly responsive to LPS (P = 0.018), as compared with healthy controls. CONCLUSIONS: Our findings reveal consistent differences in gene expression in macrophages from AS patients, with evidence of a striking reverse IFN signature. Together with poor expression and responsiveness of the IFNgamma gene, these results suggest that there may be a relative defect in IFNgamma gene regulation, with autocrine consequences and implications for disease pathogenesis.\"\n",
      "!Series_summary\t\"\"\n",
      "!Series_summary\t\"Keywords: Disease state analysis with treatment effect\"\n",
      "!Series_overall_design\t\"Macrophages were derived from the peripheral blood of 8 AS patients (median disease duration 13 years [range <1â€“43 years]) and 9 healthy control subjects over 7 days with the use of granulocyteâ€“ macrophage colony-stimulating factor. Cells were stimulated for 24 hours with interferon- (IFN; 100 units/ ml), were left untreated for 24 hours, or were treated for 3 hours with lipopolysaccharide (LPS; 10 ng/ml). RNA was isolated and examined by microarray and real-time quantitative reverse transcriptionâ€“polymerase chain reaction analysis.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM300389</th>\n",
       "      <th>GSM300390</th>\n",
       "      <th>GSM300391</th>\n",
       "      <th>GSM300392</th>\n",
       "      <th>GSM300393</th>\n",
       "      <th>GSM300394</th>\n",
       "      <th>GSM300395</th>\n",
       "      <th>GSM300396</th>\n",
       "      <th>GSM300397</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM300412</th>\n",
       "      <th>GSM300413</th>\n",
       "      <th>GSM300414</th>\n",
       "      <th>GSM300415</th>\n",
       "      <th>GSM300416</th>\n",
       "      <th>GSM300417</th>\n",
       "      <th>GSM300418</th>\n",
       "      <th>GSM300419</th>\n",
       "      <th>GSM300420</th>\n",
       "      <th>GSM300421</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Control #1, IFN treated</td>\n",
       "      <td>Control #5, IFN treated</td>\n",
       "      <td>Control #7, IFN treated</td>\n",
       "      <td>Control #2, IFN treated</td>\n",
       "      <td>Control #3, IFN treated</td>\n",
       "      <td>Control #6, IFN treated</td>\n",
       "      <td>Control #4, IFN treated</td>\n",
       "      <td>Control #8, IFN treated</td>\n",
       "      <td>Control #9, IFN treated</td>\n",
       "      <td>...</td>\n",
       "      <td>Patient #6, IFN treated</td>\n",
       "      <td>Patient #7, IFN treated</td>\n",
       "      <td>Patient #2, Untreated</td>\n",
       "      <td>Patient #8, Untreated</td>\n",
       "      <td>Patient #4, Untreated</td>\n",
       "      <td>Patient #3, Untreated</td>\n",
       "      <td>Patient #5, Untreated</td>\n",
       "      <td>Patient #1, Untreated</td>\n",
       "      <td>Patient #6, Untreated</td>\n",
       "      <td>Patient #7, Untreated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                GSM300389  \\\n",
       "0  !Sample_characteristics_ch1  Control #1, IFN treated   \n",
       "\n",
       "                 GSM300390                GSM300391                GSM300392  \\\n",
       "0  Control #5, IFN treated  Control #7, IFN treated  Control #2, IFN treated   \n",
       "\n",
       "                 GSM300393                GSM300394                GSM300395  \\\n",
       "0  Control #3, IFN treated  Control #6, IFN treated  Control #4, IFN treated   \n",
       "\n",
       "                 GSM300396                GSM300397  ...  \\\n",
       "0  Control #8, IFN treated  Control #9, IFN treated  ...   \n",
       "\n",
       "                 GSM300412                GSM300413              GSM300414  \\\n",
       "0  Patient #6, IFN treated  Patient #7, IFN treated  Patient #2, Untreated   \n",
       "\n",
       "               GSM300415              GSM300416              GSM300417  \\\n",
       "0  Patient #8, Untreated  Patient #4, Untreated  Patient #3, Untreated   \n",
       "\n",
       "               GSM300418              GSM300419              GSM300420  \\\n",
       "0  Patient #5, Untreated  Patient #1, Untreated  Patient #6, Untreated   \n",
       "\n",
       "               GSM300421  \n",
       "0  Patient #7, Untreated  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: How to convert trait? (control vs patient? IFN vs Do nothing?)\n",
    "cohort = accession_num = \"GSE11886\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'Control #1, IFN treated',\n",
       "       'Control #5, IFN treated', 'Control #7, IFN treated',\n",
       "       'Control #2, IFN treated', 'Control #3, IFN treated',\n",
       "       'Control #6, IFN treated', 'Control #4, IFN treated',\n",
       "       'Control #8, IFN treated', 'Control #9, IFN treated',\n",
       "       'Control #1, Untreated', 'Control #5, Untreated',\n",
       "       'Control #7, Untreated', 'Control #2, Untreated',\n",
       "       'Control #3, Untreated', 'Control #6, Untreated',\n",
       "       'Control #4, Untreated', 'Control #8, Untreated',\n",
       "       'Control #9, Untreated', 'Patient #2, Untreated',\n",
       "       'Patient 42, IFN treated', 'Patient #3, IFN treated',\n",
       "       'Patient #5, IFN treated', 'Patient #1, IFN treated',\n",
       "       'Patient #6, IFN treated', 'Patient #7, IFN treated',\n",
       "       'Patient #8, Untreated', 'Patient #4, Untreated',\n",
       "       'Patient #3, Untreated', 'Patient #5, Untreated',\n",
       "       'Patient #1, Untreated', 'Patient #6, Untreated',\n",
       "       'Patient #7, Untreated'], dtype=object)"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression analysis of macrophages from ankylosing spondylitis patients reveals interferon-gamma dysregulation\"\n",
      "!Series_summary\t\"OBJECTIVE: To determine whether macrophages, a type of cell implicated in the pathogenesis of ankylosing spondylitis (AS), exhibit a characteristic gene expression pattern. METHODS: Macrophages were derived from the peripheral blood of 8 AS patients (median disease duration 13 years [range <1-43 years]) and 9 healthy control subjects over 7 days with the use of granulocyte-macrophage colony-stimulating factor. Cells were stimulated for 24 hours with interferon-gamma (IFNgamma; 100 units/ml), were left untreated for 24 hours, or were treated for 3 hours with lipopolysaccharide (LPS; 10 ng/ml). RNA was isolated and examined by microarray and real-time quantitative reverse transcription-polymerase chain reaction analysis. RESULTS: Microarray analysis revealed 198 probe sets detecting the differential expression of 141 unique genes in untreated macrophages from AS patients compared with healthy controls. Clustering and principal components analysis clearly distinguished AS patients and controls. Of the differentially expressed genes, 78 (55%) were IFN-regulated, and their relative expression indicated a reverse IFN signature in AS patient macrophages, where IFNgamma-up-regulated genes were underexpressed and down-regulated genes were overexpressed. Treatment of macrophages with exogenous IFNgamma normalized the expression of these genes between patients and controls. In addition, the messenger RNA encoded by the IFNgamma gene was approximately 2-fold lower in AS patient macrophages at baseline (P = 0.004) and was poorly responsive to LPS (P = 0.018), as compared with healthy controls. CONCLUSIONS: Our findings reveal consistent differences in gene expression in macrophages from AS patients, with evidence of a striking reverse IFN signature. Together with poor expression and responsiveness of the IFNgamma gene, these results suggest that there may be a relative defect in IFNgamma gene regulation, with autocrine consequences and implications for disease pathogenesis.\"\n",
      "!Series_summary\t\"\"\n",
      "!Series_summary\t\"Keywords: Disease state analysis with treatment effect\"\n",
      "!Series_overall_design\t\"Macrophages were derived from the peripheral blood of 8 AS patients (median disease duration 13 years [range <1â€“43 years]) and 9 healthy control subjects over 7 days with the use of granulocyteâ€“ macrophage colony-stimulating factor. Cells were stimulated for 24 hours with interferon- (IFN; 100 units/ ml), were left untreated for 24 hours, or were treated for 3 hours with lipopolysaccharide (LPS; 10 ng/ml). RNA was isolated and examined by microarray and real-time quantitative reverse transcriptionâ€“polymerase chain reaction analysis.\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM300389</th>\n",
       "      <th>GSM300390</th>\n",
       "      <th>GSM300391</th>\n",
       "      <th>GSM300392</th>\n",
       "      <th>GSM300393</th>\n",
       "      <th>GSM300394</th>\n",
       "      <th>GSM300395</th>\n",
       "      <th>GSM300396</th>\n",
       "      <th>GSM300397</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM300412</th>\n",
       "      <th>GSM300413</th>\n",
       "      <th>GSM300414</th>\n",
       "      <th>GSM300415</th>\n",
       "      <th>GSM300416</th>\n",
       "      <th>GSM300417</th>\n",
       "      <th>GSM300418</th>\n",
       "      <th>GSM300419</th>\n",
       "      <th>GSM300420</th>\n",
       "      <th>GSM300421</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Control #1, IFN treated</td>\n",
       "      <td>Control #5, IFN treated</td>\n",
       "      <td>Control #7, IFN treated</td>\n",
       "      <td>Control #2, IFN treated</td>\n",
       "      <td>Control #3, IFN treated</td>\n",
       "      <td>Control #6, IFN treated</td>\n",
       "      <td>Control #4, IFN treated</td>\n",
       "      <td>Control #8, IFN treated</td>\n",
       "      <td>Control #9, IFN treated</td>\n",
       "      <td>...</td>\n",
       "      <td>Patient #6, IFN treated</td>\n",
       "      <td>Patient #7, IFN treated</td>\n",
       "      <td>Patient #2, Untreated</td>\n",
       "      <td>Patient #8, Untreated</td>\n",
       "      <td>Patient #4, Untreated</td>\n",
       "      <td>Patient #3, Untreated</td>\n",
       "      <td>Patient #5, Untreated</td>\n",
       "      <td>Patient #1, Untreated</td>\n",
       "      <td>Patient #6, Untreated</td>\n",
       "      <td>Patient #7, Untreated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                GSM300389  \\\n",
       "0  !Sample_characteristics_ch1  Control #1, IFN treated   \n",
       "\n",
       "                 GSM300390                GSM300391                GSM300392  \\\n",
       "0  Control #5, IFN treated  Control #7, IFN treated  Control #2, IFN treated   \n",
       "\n",
       "                 GSM300393                GSM300394                GSM300395  \\\n",
       "0  Control #3, IFN treated  Control #6, IFN treated  Control #4, IFN treated   \n",
       "\n",
       "                 GSM300396                GSM300397  ...  \\\n",
       "0  Control #8, IFN treated  Control #9, IFN treated  ...   \n",
       "\n",
       "                 GSM300412                GSM300413              GSM300414  \\\n",
       "0  Patient #6, IFN treated  Patient #7, IFN treated  Patient #2, Untreated   \n",
       "\n",
       "               GSM300415              GSM300416              GSM300417  \\\n",
       "0  Patient #8, Untreated  Patient #4, Untreated  Patient #3, Untreated   \n",
       "\n",
       "               GSM300418              GSM300419              GSM300420  \\\n",
       "0  Patient #5, Untreated  Patient #1, Untreated  Patient #6, Untreated   \n",
       "\n",
       "               GSM300421  \n",
       "0  Patient #7, Untreated  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['Control #1, IFN treated',\n",
       "  'Control #5, IFN treated',\n",
       "  'Control #7, IFN treated',\n",
       "  'Control #2, IFN treated',\n",
       "  'Control #3, IFN treated',\n",
       "  'Control #6, IFN treated',\n",
       "  'Control #4, IFN treated',\n",
       "  'Control #8, IFN treated',\n",
       "  'Control #9, IFN treated',\n",
       "  'Control #1, Untreated',\n",
       "  'Control #5, Untreated',\n",
       "  'Control #7, Untreated',\n",
       "  'Control #2, Untreated',\n",
       "  'Control #3, Untreated',\n",
       "  'Control #6, Untreated',\n",
       "  'Control #4, Untreated',\n",
       "  'Control #8, Untreated',\n",
       "  'Control #9, Untreated',\n",
       "  'Patient #2, Untreated',\n",
       "  'Patient 42, IFN treated',\n",
       "  'Patient #3, IFN treated',\n",
       "  'Patient #5, IFN treated',\n",
       "  'Patient #1, IFN treated',\n",
       "  'Patient #6, IFN treated',\n",
       "  'Patient #7, IFN treated',\n",
       "  'Patient #8, Untreated',\n",
       "  'Patient #4, Untreated',\n",
       "  'Patient #3, Untreated',\n",
       "  'Patient #5, Untreated',\n",
       "  'Patient #1, Untreated']}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Ankylosing Spondylitis\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Ankylosing Spondylitis\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Gene expression analysis of macrophages from ankylosing spondylitis patients reveals interferon-gamma dysregulation\"\\n!Series_summary\\t\"OBJECTIVE: To determine whether macrophages, a type of cell implicated in the pathogenesis of ankylosing spondylitis (AS), exhibit a characteristic gene expression pattern. METHODS: Macrophages were derived from the peripheral blood of 8 AS patients (median disease duration 13 years [range <1-43 years]) and 9 healthy control subjects over 7 days with the use of granulocyte-macrophage colony-stimulating factor. Cells were stimulated for 24 hours with interferon-gamma (IFNgamma; 100 units/ml), were left untreated for 24 hours, or were treated for 3 hours with lipopolysaccharide (LPS; 10 ng/ml). RNA was isolated and examined by microarray and real-time quantitative reverse transcription-polymerase chain reaction analysis. RESULTS: Microarray analysis revealed 198 probe sets detecting the differential expression of 141 unique genes in untreated macrophages from AS patients compared with healthy controls. Clustering and principal components analysis clearly distinguished AS patients and controls. Of the differentially expressed genes, 78 (55%) were IFN-regulated, and their relative expression indicated a reverse IFN signature in AS patient macrophages, where IFNgamma-up-regulated genes were underexpressed and down-regulated genes were overexpressed. Treatment of macrophages with exogenous IFNgamma normalized the expression of these genes between patients and controls. In addition, the messenger RNA encoded by the IFNgamma gene was approximately 2-fold lower in AS patient macrophages at baseline (P = 0.004) and was poorly responsive to LPS (P = 0.018), as compared with healthy controls. CONCLUSIONS: Our findings reveal consistent differences in gene expression in macrophages from AS patients, with evidence of a striking reverse IFN signature. Together with poor expression and responsiveness of the IFNgamma gene, these results suggest that there may be a relative defect in IFNgamma gene regulation, with autocrine consequences and implications for disease pathogenesis.\"\\n!Series_summary\\t\"\"\\n!Series_summary\\t\"Keywords: Disease state analysis with treatment effect\"\\n!Series_overall_design\\t\"Macrophages were derived from the peripheral blood of 8 AS patients (median disease duration 13 years [range <1â€“43 years]) and 9 healthy control subjects over 7 days with the use of granulocyteâ€“ macrophage colony-stimulating factor. Cells were stimulated for 24 hours with interferon- (IFN; 100 units/ ml), were left untreated for 24 hours, or were treated for 3 hours with lipopolysaccharide (LPS; 10 ng/ml). RNA was isolated and examined by microarray and real-time quantitative reverse transcriptionâ€“polymerase chain reaction analysis.\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'Control #1, IFN treated\\', \\'Control #5, IFN treated\\', \\'Control #7, IFN treated\\', \\'Control #2, IFN treated\\', \\'Control #3, IFN treated\\', \\'Control #6, IFN treated\\', \\'Control #4, IFN treated\\', \\'Control #8, IFN treated\\', \\'Control #9, IFN treated\\', \\'Control #1, Untreated\\', \\'Control #5, Untreated\\', \\'Control #7, Untreated\\', \\'Control #2, Untreated\\', \\'Control #3, Untreated\\', \\'Control #6, Untreated\\', \\'Control #4, Untreated\\', \\'Control #8, Untreated\\', \\'Control #9, Untreated\\', \\'Patient #2, Untreated\\', \\'Patient 42, IFN treated\\', \\'Patient #3, IFN treated\\', \\'Patient #5, IFN treated\\', \\'Patient #1, IFN treated\\', \\'Patient #6, IFN treated\\', \\'Patient #7, IFN treated\\', \\'Patient #8, Untreated\\', \\'Patient #4, Untreated\\', \\'Patient #3, Untreated\\', \\'Patient #5, Untreated\\', \\'Patient #1, Untreated\\']}\\n'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'Control #1, IFN treated' or tissue_type == 'Control #5, IFN treated'or tissue_type == 'Control #7, IFN treated'or tissue_type == 'Control #2, IFN treated'or tissue_type == 'Control #3, IFN treated'or tissue_type == 'Control #6, IFN treated'or tissue_type == 'Control #4, IFN treated'or tissue_type == 'Control #8, IFN treated'or tissue_type == 'Control #9, IFN treated'or tissue_type == 'Control 42, IFN treated':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if gender_string.lower() == 'sex: female':\n",
    "        return 1\n",
    "    elif gender_string.lower() == 'sex: male':\n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_24124\\2239879167.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM300389</th>\n",
       "      <th>GSM300390</th>\n",
       "      <th>GSM300391</th>\n",
       "      <th>GSM300392</th>\n",
       "      <th>GSM300393</th>\n",
       "      <th>GSM300394</th>\n",
       "      <th>GSM300395</th>\n",
       "      <th>GSM300396</th>\n",
       "      <th>GSM300397</th>\n",
       "      <th>GSM300398</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM300412</th>\n",
       "      <th>GSM300413</th>\n",
       "      <th>GSM300414</th>\n",
       "      <th>GSM300415</th>\n",
       "      <th>GSM300416</th>\n",
       "      <th>GSM300417</th>\n",
       "      <th>GSM300418</th>\n",
       "      <th>GSM300419</th>\n",
       "      <th>GSM300420</th>\n",
       "      <th>GSM300421</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ankylosing Spondylitis</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GSM300389  GSM300390  GSM300391  GSM300392  GSM300393  \\\n",
       "Ankylosing Spondylitis          1          1          1          1          1   \n",
       "\n",
       "                        GSM300394  GSM300395  GSM300396  GSM300397  GSM300398  \\\n",
       "Ankylosing Spondylitis          1          1          1          1          0   \n",
       "\n",
       "                        ...  GSM300412  GSM300413  GSM300414  GSM300415  \\\n",
       "Ankylosing Spondylitis  ...          0          0          0          0   \n",
       "\n",
       "                        GSM300416  GSM300417  GSM300418  GSM300419  GSM300420  \\\n",
       "Ankylosing Spondylitis          0          0          0          0          0   \n",
       "\n",
       "                        GSM300421  \n",
       "Ankylosing Spondylitis          0  \n",
       "\n",
       "[1 rows x 33 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM300389</th>\n",
       "      <th>GSM300390</th>\n",
       "      <th>GSM300391</th>\n",
       "      <th>GSM300392</th>\n",
       "      <th>GSM300393</th>\n",
       "      <th>GSM300394</th>\n",
       "      <th>GSM300395</th>\n",
       "      <th>GSM300396</th>\n",
       "      <th>GSM300397</th>\n",
       "      <th>GSM300398</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM300412</th>\n",
       "      <th>GSM300413</th>\n",
       "      <th>GSM300414</th>\n",
       "      <th>GSM300415</th>\n",
       "      <th>GSM300416</th>\n",
       "      <th>GSM300417</th>\n",
       "      <th>GSM300418</th>\n",
       "      <th>GSM300419</th>\n",
       "      <th>GSM300420</th>\n",
       "      <th>GSM300421</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>0.315248</td>\n",
       "      <td>0.112150</td>\n",
       "      <td>0.070344</td>\n",
       "      <td>0.248315</td>\n",
       "      <td>0.020277</td>\n",
       "      <td>-0.150602</td>\n",
       "      <td>0.060186</td>\n",
       "      <td>-0.001738</td>\n",
       "      <td>0.139716</td>\n",
       "      <td>0.103189</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076530</td>\n",
       "      <td>-0.004617</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.154255</td>\n",
       "      <td>-0.010868</td>\n",
       "      <td>-0.270209</td>\n",
       "      <td>-0.055730</td>\n",
       "      <td>0.071297</td>\n",
       "      <td>0.110618</td>\n",
       "      <td>-0.171019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>-0.389087</td>\n",
       "      <td>-0.814719</td>\n",
       "      <td>-0.879279</td>\n",
       "      <td>-0.310306</td>\n",
       "      <td>-0.975432</td>\n",
       "      <td>-0.997373</td>\n",
       "      <td>-0.852638</td>\n",
       "      <td>-0.490865</td>\n",
       "      <td>-0.105638</td>\n",
       "      <td>0.232449</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.721043</td>\n",
       "      <td>-0.932037</td>\n",
       "      <td>0.576449</td>\n",
       "      <td>0.601156</td>\n",
       "      <td>0.393908</td>\n",
       "      <td>0.478886</td>\n",
       "      <td>-0.259640</td>\n",
       "      <td>-0.210929</td>\n",
       "      <td>-0.120008</td>\n",
       "      <td>0.164282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>0.080380</td>\n",
       "      <td>0.006952</td>\n",
       "      <td>0.128363</td>\n",
       "      <td>-0.112075</td>\n",
       "      <td>0.272227</td>\n",
       "      <td>0.141575</td>\n",
       "      <td>0.937242</td>\n",
       "      <td>0.018661</td>\n",
       "      <td>0.091574</td>\n",
       "      <td>-0.058224</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.159909</td>\n",
       "      <td>0.021100</td>\n",
       "      <td>0.034181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004055</td>\n",
       "      <td>-0.166020</td>\n",
       "      <td>-0.105684</td>\n",
       "      <td>-0.039442</td>\n",
       "      <td>-0.310866</td>\n",
       "      <td>-0.257435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>0.159040</td>\n",
       "      <td>-0.029855</td>\n",
       "      <td>-0.065653</td>\n",
       "      <td>-0.014381</td>\n",
       "      <td>-0.015624</td>\n",
       "      <td>0.008447</td>\n",
       "      <td>-0.228561</td>\n",
       "      <td>-0.146182</td>\n",
       "      <td>0.123858</td>\n",
       "      <td>0.208845</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.204799</td>\n",
       "      <td>-0.251754</td>\n",
       "      <td>-0.095480</td>\n",
       "      <td>0.245110</td>\n",
       "      <td>0.043694</td>\n",
       "      <td>-0.265528</td>\n",
       "      <td>0.124260</td>\n",
       "      <td>-0.075810</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.365727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>0.240606</td>\n",
       "      <td>0.073325</td>\n",
       "      <td>0.142995</td>\n",
       "      <td>0.058000</td>\n",
       "      <td>0.006382</td>\n",
       "      <td>0.239335</td>\n",
       "      <td>0.062448</td>\n",
       "      <td>0.221553</td>\n",
       "      <td>0.279908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.003346</td>\n",
       "      <td>0.267826</td>\n",
       "      <td>0.149425</td>\n",
       "      <td>0.106815</td>\n",
       "      <td>0.047712</td>\n",
       "      <td>0.035424</td>\n",
       "      <td>0.042320</td>\n",
       "      <td>-0.025099</td>\n",
       "      <td>-0.018436</td>\n",
       "      <td>-0.038872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM300389  GSM300390  GSM300391  GSM300392  GSM300393  GSM300394  \\\n",
       "ID                                                                            \n",
       "1007_s_at   0.315248   0.112150   0.070344   0.248315   0.020277  -0.150602   \n",
       "1053_at    -0.389087  -0.814719  -0.879279  -0.310306  -0.975432  -0.997373   \n",
       "117_at      0.080380   0.006952   0.128363  -0.112075   0.272227   0.141575   \n",
       "121_at      0.159040  -0.029855  -0.065653  -0.014381  -0.015624   0.008447   \n",
       "1255_g_at   0.240606   0.073325   0.142995   0.058000   0.006382   0.239335   \n",
       "\n",
       "           GSM300395  GSM300396  GSM300397  GSM300398  ...  GSM300412  \\\n",
       "ID                                                     ...              \n",
       "1007_s_at   0.060186  -0.001738   0.139716   0.103189  ...   0.076530   \n",
       "1053_at    -0.852638  -0.490865  -0.105638   0.232449  ...  -0.721043   \n",
       "117_at      0.937242   0.018661   0.091574  -0.058224  ...  -0.159909   \n",
       "121_at     -0.228561  -0.146182   0.123858   0.208845  ...  -0.204799   \n",
       "1255_g_at   0.062448   0.221553   0.279908   0.000000  ...  -0.003346   \n",
       "\n",
       "           GSM300413  GSM300414  GSM300415  GSM300416  GSM300417  GSM300418  \\\n",
       "ID                                                                            \n",
       "1007_s_at  -0.004617   0.000000   0.154255  -0.010868  -0.270209  -0.055730   \n",
       "1053_at    -0.932037   0.576449   0.601156   0.393908   0.478886  -0.259640   \n",
       "117_at      0.021100   0.034181   0.000000   0.004055  -0.166020  -0.105684   \n",
       "121_at     -0.251754  -0.095480   0.245110   0.043694  -0.265528   0.124260   \n",
       "1255_g_at   0.267826   0.149425   0.106815   0.047712   0.035424   0.042320   \n",
       "\n",
       "           GSM300419  GSM300420  GSM300421  \n",
       "ID                                          \n",
       "1007_s_at   0.071297   0.110618  -0.171019  \n",
       "1053_at    -0.210929  -0.120008   0.164282  \n",
       "117_at     -0.039442  -0.310866  -0.257435  \n",
       "121_at     -0.075810   0.000000  -0.365727  \n",
       "1255_g_at  -0.025099  -0.018436  -0.038872  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1007_s_at',\n",
       " '1053_at',\n",
       " '117_at',\n",
       " '121_at',\n",
       " '1255_g_at',\n",
       " '1294_at',\n",
       " '1316_at',\n",
       " '1320_at',\n",
       " '1405_i_at',\n",
       " '1431_at',\n",
       " '1438_at',\n",
       " '1487_at',\n",
       " '1494_f_at',\n",
       " '1552256_a_at',\n",
       " '1552257_a_at',\n",
       " '1552258_at',\n",
       " '1552261_at',\n",
       " '1552263_at',\n",
       " '1552264_a_at',\n",
       " '1552266_at']"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\\n\""
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'UCSC_RefGene_Name'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_a\n",
      "1000 input query terms found no hit:\t['1553625_at', '1553626_a_at', '1553627_s_at', '1553629_a_at', '1553630_at', '1553633_s_at', '155363\n",
      "1000 input query terms found no hit:\t['1555009_a_at', '1555011_at', '1555014_x_at', '1555015_a_at', '1555016_at', '1555018_at', '1555019_\n",
      "1000 input query terms found no hit:\t['1556425_a_at', '1556426_at', '1556427_s_at', '1556429_a_at', '1556432_at', '1556434_at', '1556435_\n",
      "1000 input query terms found no hit:\t['1557981_at', '1557984_s_at', '1557985_s_at', '1557986_s_at', '1557987_at', '1557991_at', '1557993_\n",
      "1000 input query terms found no hit:\t['1559795_at', '1559796_at', '1559800_a_at', '1559804_at', '1559806_at', '1559807_at', '1559808_at',\n",
      "1000 input query terms found no hit:\t['1561429_a_at', '1561430_s_at', '1561431_at', '1561432_at', '1561433_at', '1561434_at', '1561436_at\n",
      "1000 input query terms found no hit:\t['1563054_at', '1563055_at', '1563057_at', '1563059_at', '1563061_at', '1563062_at', '1563063_at', '\n",
      "1000 input query terms found no hit:\t['1565787_at', '1565788_at', '1565795_at', '1565797_at', '1565799_at', '1565800_x_at', '1565801_at',\n",
      "1000 input query terms found no hit:\t['1569042_at', '1569044_at', '1569048_s_at', '1569052_at', '1569053_at', '1569054_at', '1569057_s_at\n",
      "1000 input query terms found no hit:\t['200059_s_at', '200060_s_at', '200061_s_at', '200062_s_at', '200063_s_at', '200064_at', '200065_s_a\n",
      "1000 input query terms found no hit:\t['201552_at', '201553_s_at', '201554_x_at', '201555_at', '201556_s_at', '201557_at', '201558_at', '2\n",
      "1000 input query terms found no hit:\t['202552_s_at', '202553_s_at', '202554_s_at', '202555_s_at', '202556_s_at', '202557_at', '202558_s_a\n",
      "1000 input query terms found no hit:\t['203553_s_at', '203554_x_at', '203555_at', '203556_at', '203557_s_at', '203558_at', '203559_s_at', \n",
      "1000 input query terms found no hit:\t['204553_x_at', '204554_at', '204555_s_at', '204556_s_at', '204557_s_at', '204558_at', '204559_s_at'\n",
      "1000 input query terms found no hit:\t['205553_s_at', '205554_s_at', '205555_s_at', '205556_at', '205557_at', '205558_at', '205559_s_at', \n",
      "1000 input query terms found no hit:\t['206554_x_at', '206555_s_at', '206556_at', '206557_at', '206558_at', '206559_x_at', '206560_s_at', \n",
      "1000 input query terms found no hit:\t['207559_s_at', '207560_at', '207561_s_at', '207562_at', '207563_s_at', '207564_x_at', '207565_s_at'\n",
      "1000 input query terms found no hit:\t['208582_s_at', '208583_x_at', '208584_at', '208585_at', '208586_s_at', '208587_s_at', '208588_at', \n",
      "1000 input query terms found no hit:\t['209590_at', '209591_s_at', '209592_s_at', '209593_s_at', '209594_x_at', '209595_at', '209596_at', \n",
      "1000 input query terms found no hit:\t['210615_at', '210616_s_at', '210617_at', '210618_at', '210619_s_at', '210620_s_at', '210621_s_at', \n",
      "1000 input query terms found no hit:\t['211678_s_at', '211679_x_at', '211680_at', '211681_s_at', '211682_x_at', '211684_s_at', '211685_s_a\n",
      "1000 input query terms found no hit:\t['212695_at', '212696_s_at', '212697_at', '212698_s_at', '212699_at', '212700_x_at', '212701_at', '2\n",
      "1000 input query terms found no hit:\t['213700_s_at', '213701_at', '213702_x_at', '213703_at', '213704_at', '213705_at', '213706_at', '213\n",
      "1000 input query terms found no hit:\t['214704_at', '214705_at', '214706_at', '214707_x_at', '214708_at', '214709_s_at', '214710_s_at', '2\n",
      "1000 input query terms found no hit:\t['215707_s_at', '215708_s_at', '215709_at', '215710_at', '215711_s_at', '215712_s_at', '215713_at', \n",
      "1000 input query terms found no hit:\t['216710_x_at', '216711_s_at', '216712_at', '216713_at', '216714_at', '216715_at', '216716_at', '216\n",
      "1000 input query terms found no hit:\t['217715_x_at', '217716_s_at', '217717_s_at', '217718_s_at', '217719_at', '217720_at', '217721_at', \n",
      "1000 input query terms found no hit:\t['218716_x_at', '218717_s_at', '218718_at', '218719_s_at', '218720_x_at', '218721_s_at', '218722_s_a\n",
      "1000 input query terms found no hit:\t['219716_at', '219717_at', '219718_at', '219719_at', '219720_s_at', '219721_at', '219722_s_at', '219\n",
      "1000 input query terms found no hit:\t['220716_at', '220717_at', '220718_at', '220719_at', '220720_x_at', '220721_at', '220722_s_at', '220\n",
      "1000 input query terms found no hit:\t['221720_s_at', '221721_s_at', '221722_x_at', '221723_s_at', '221724_s_at', '221725_at', '221726_at'\n",
      "1000 input query terms found no hit:\t['222721_at', '222722_at', '222723_at', '222724_at', '222725_s_at', '222726_s_at', '222727_s_at', '2\n",
      "1000 input query terms found no hit:\t['223724_s_at', '223725_at', '223726_at', '223727_at', '223728_at', '223729_at', '223730_at', '22373\n",
      "1000 input query terms found no hit:\t['224741_x_at', '224742_at', '224743_at', '224744_at', '224745_x_at', '224746_at', '224747_at', '224\n",
      "1000 input query terms found no hit:\t['225743_at', '225744_at', '225745_at', '225746_at', '225747_at', '225748_at', '225749_at', '225750_\n",
      "1000 input query terms found no hit:\t['226745_at', '226746_s_at', '226747_at', '226748_at', '226749_at', '226750_at', '226751_at', '22675\n",
      "1000 input query terms found no hit:\t['227746_at', '227747_at', '227748_at', '227749_at', '227750_at', '227751_at', '227752_at', '227753_\n",
      "1000 input query terms found no hit:\t['228746_s_at', '228747_at', '228748_at', '228749_at', '228750_at', '228751_at', '228752_at', '22875\n",
      "1000 input query terms found no hit:\t['229746_x_at', '229747_x_at', '229748_x_at', '229749_at', '229750_at', '229751_s_at', '229752_at', \n",
      "1000 input query terms found no hit:\t['230746_s_at', '230747_s_at', '230748_at', '230749_s_at', '230750_at', '230751_at', '230752_at', '2\n",
      "1000 input query terms found no hit:\t['231746_at', '231747_at', '231748_at', '231749_at', '231750_at', '231751_at', '231752_at', '231753_\n",
      "1000 input query terms found no hit:\t['232747_at', '232748_at', '232749_at', '232750_at', '232751_at', '232752_at', '232753_at', '232754_\n",
      "1000 input query terms found no hit:\t['233749_at', '233750_s_at', '233751_at', '233752_s_at', '233753_at', '233754_x_at', '233755_at', '2\n",
      "1000 input query terms found no hit:\t['234751_s_at', '234752_x_at', '234753_x_at', '234754_at', '234755_x_at', '234756_at', '234757_at', \n",
      "1000 input query terms found no hit:\t['235751_s_at', '235752_at', '235753_at', '235754_at', '235755_at', '235756_at', '235757_at', '23575\n",
      "1000 input query terms found no hit:\t['236751_at', '236752_at', '236753_at', '236754_at', '236755_at', '236756_at', '236757_at', '236758_\n",
      "1000 input query terms found no hit:\t['237751_x_at', '237752_at', '237753_at', '237754_at', '237755_s_at', '237756_at', '237757_at', '237\n",
      "1000 input query terms found no hit:\t['238751_at', '238752_at', '238753_at', '238754_at', '238755_at', '238756_at', '238757_at', '238758_\n",
      "1000 input query terms found no hit:\t['239751_at', '239752_at', '239753_at', '239754_at', '239755_at', '239756_at', '239757_at', '239758_\n",
      "1000 input query terms found no hit:\t['240751_at', '240752_at', '240753_at', '240754_at', '240755_at', '240756_at', '240757_at', '240758_\n",
      "1000 input query terms found no hit:\t['241751_at', '241752_at', '241753_at', '241754_at', '241755_at', '241756_at', '241757_x_at', '24175\n",
      "1000 input query terms found no hit:\t['242751_at', '242752_at', '242753_x_at', '242754_at', '242755_at', '242756_at', '242757_at', '24275\n",
      "1000 input query terms found no hit:\t['243751_at', '243752_s_at', '243753_at', '243754_at', '243755_at', '243756_at', '243757_at', '24375\n",
      "675 input query terms found no hit:\t['244750_at', '244751_at', '244752_at', '244753_at', '244754_at', '244755_at', '244756_at', '244757_\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 33 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Ankylosing Spondylitis', the least common label is '1.0' with 9 occurrences. This represents 27.27% of the dataset.\n",
      "The distribution of the feature 'Ankylosing Spondylitis' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE11886</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>33</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "1  GSE11886       True          True      False    False       False   \n",
       "\n",
       "   sample_size note  \n",
       "1           33       "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.295070023Z",
     "start_time": "2024-01-10T21:40:21.338515691Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +: 'NoneType' and 'str'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[84], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m merged_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, \u001b[43mcondition_best_cohort\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m))\n\u001b[0;32m      2\u001b[0m merged_data\u001b[38;5;241m.\u001b[39mhead()\n",
      "\u001b[1;31mTypeError\u001b[0m: unsupported operand type(s) for +: 'NoneType' and 'str'"
     ]
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.358779256Z",
     "start_time": "2024-01-10T21:40:27.271337819Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n",
    "\n",
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546590575Z",
     "start_time": "2024-01-10T21:40:27.298495045Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546733889Z",
     "start_time": "2024-01-10T21:40:27.535249532Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Select appropriate models based on whether the dataset has batch effect.\n",
    "# We experiment on two models for each branch. We will decide which one to choose later.\n",
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:30.570471636Z",
     "start_time": "2024-01-10T21:40:27.546295159Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:47.575878022Z",
     "start_time": "2024-01-10T21:40:30.570907992Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.663737008Z",
     "start_time": "2024-01-10T21:40:47.576021186Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.673565176Z",
     "start_time": "2024-01-10T21:40:51.665291317Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.730825913Z",
     "start_time": "2024-01-10T21:40:51.675049728Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
