{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "# Set your preferred name\n",
    "USER = \"Jiayi\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output'\n",
    "TRAIT = 'Anxiety disorder'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.928143012Z",
     "start_time": "2024-01-10T21:37:25.918685567Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.554597756Z",
     "start_time": "2024-01-10T21:37:26.530982646Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.999521704Z",
     "start_time": "2024-01-10T21:37:26.987471072Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Adrenocortical_Cancer_(ACC)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_ACC_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.180581225Z",
     "start_time": "2024-01-10T21:37:27.217363685Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.185775284Z",
     "start_time": "2024-01-10T21:37:30.181745846Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'additional_pharmaceutical_therapy',\n",
       " 'additional_radiation_therapy',\n",
       " 'age_at_initial_pathologic_diagnosis',\n",
       " 'atypical_mitotic_figures',\n",
       " 'bcr_followup_barcode']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.269854441Z",
     "start_time": "2024-01-10T21:37:30.184225525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'age_at_initial_pathologic_diagnosis', 'atypical_mitotic_figures', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'clinical_M', 'ct_scan_findings', 'cytoplasm_presence_less_than_equal_25_percent', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'diffuse_architecture', 'distant_metastasis_anatomic_site', 'excess_adrenal_hormone_diagnosis_method_type', 'excess_adrenal_hormone_history_type', 'form_completion_date', 'gender', 'germline_testing_performed', 'histologic_disease_progression_present_indicator', 'histological_type', 'history_of_neoadjuvant_treatment', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'initial_weight', 'invasion_of_tumor_capsule', 'is_ffpe', 'laterality', 'lost_follow_up', 'lymph_node_examined_count', 'metastatic_neoplasm_confirmed_diagnosis_method_name', 'metastatic_neoplasm_confirmed_diagnosis_method_text', 'mitoses_count', 'mitotane_therapy', 'mitotane_therapy_adjuvant_setting', 'mitotane_therapy_for_macroscopic_residual_disease', 'mitotic_rate', 'necrosis', 'new_neoplasm_confirmed_diagnosis_method_name', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'nuclear_grade_III_IV', 'number_of_lymphnodes_positive_by_he', 'oct_embedded', 'other_dx', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'person_neoplasm_cancer_status', 'post_surgical_procedure_assessment_thyroid_gland_carcinoma_stats', 'postoperative_rx_tx', 'primary_lymph_node_presentation_assessment', 'primary_therapy_outcome_success', 'radiation_therapy', 'residual_tumor', 'ret', 'sample_type', 'sample_type_id', 'sinusoid_invasion', 'therapeutic_mitotane_levels_achieved', 'therapeutic_mitotane_lvl_macroscopic_residual', 'therapeutic_mitotane_lvl_progression', 'therapeutic_mitotane_lvl_recurrence', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'vial_number', 'vital_status', 'weiss_score', 'weiss_venous_invasion', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_ACC_mutation_curated_bcm_gene', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_percentile', '_GENOMIC_ID_data/public/TCGA/ACC/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_ACC_miRNA_HiSeq', '_GENOMIC_ID_TCGA_ACC_RPPA', '_GENOMIC_ID_TCGA_ACC_hMethyl450', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2', '_GENOMIC_ID_TCGA_ACC_gistic2thd', '_GENOMIC_ID_TCGA_ACC_PDMRNAseq', '_GENOMIC_ID_TCGA_ACC_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_ACC_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_ACC_gistic2', '_GENOMIC_ID_TCGA_ACC_mutation_bcm_gene', '_GENOMIC_ID_TCGA_ACC_mutation_curated_broad_gene']\\n\""
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270036796Z",
     "start_time": "2024-01-10T21:37:30.227180075Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = [ 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = [ 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270232386Z",
     "start_time": "2024-01-10T21:37:30.227282564Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'age_at_initial_pathologic_diagnosis': [58, 44, 23, 23, 30],\n",
       " 'days_to_birth': [-21496, -16090, -8624, -8451, -11171],\n",
       " 'year_of_initial_pathologic_diagnosis': [2000, 2004, 2008, 2000, 2000]}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270339062Z",
     "start_time": "2024-01-10T21:37:30.227412883Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270520195Z",
     "start_time": "2024-01-10T21:37:30.227467704Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gender': ['MALE', 'FEMALE', 'FEMALE', 'FEMALE', 'MALE']}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270626721Z",
     "start_time": "2024-01-10T21:37:30.227530279Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270719762Z",
     "start_time": "2024-01-10T21:37:30.227585330Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.001977877Z",
     "start_time": "2024-01-10T21:37:30.227635342Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12 input query terms found dup hits:\t[('GTF2IP1', 2), ('RBMY1A3P', 3), ('RPL31P11', 2), ('HERC2P2', 3), ('WASH3P', 3), ('NUDT9P1', 2), ('\n",
      "154 input query terms found no hit:\t['C16orf13', 'C16orf11', 'LOC100272146', 'LOC339240', 'NACAP1', 'LOC441204', 'KLRA1', 'FAM183A', 'FA\n",
      "10 input query terms found dup hits:\t[('SUGT1P1', 2), ('PTPRVP', 2), ('SNORA62', 3), ('IFITM4P', 7), ('HLA-DRB6', 2), ('FUNDC2P2', 2), ('\n",
      "190 input query terms found no hit:\t['NARFL', 'NFKBIL2', 'LOC150197', 'TMEM84', 'LOC162632', 'PPPDE1', 'PPPDE2', 'C1orf38', 'C1orf31', '\n",
      "11 input query terms found dup hits:\t[('PIP5K1P1', 2), ('HBD', 2), ('PPP1R2P1', 9), ('HSD17B7P2', 2), ('RPSAP9', 2), ('SNORD68', 2), ('SN\n",
      "149 input query terms found no hit:\t['FAM153C', 'C9orf167', 'CLK2P', 'CCDC76', 'CCDC75', 'CCDC72', 'HIST3H2BB', 'PRAC', 'LOC285780', 'LO\n",
      "15 input query terms found dup hits:\t[('SNORD58C', 2), ('UOX', 2), ('UBE2Q2P1', 3), ('PPP4R1L', 2), ('SNORD63', 3), ('ESPNP', 2), ('HBBP1\n",
      "158 input query terms found no hit:\t['C16orf91', 'LOC100130581', 'SFRS2', 'C16orf5', 'C16orf7', 'C16orf3', 'C21orf7', 'C21orf2', 'LOC389\n",
      "15 input query terms found dup hits:\t[('FAM66D', 3), ('FAM66A', 2), ('THSD1P1', 2), ('EEF1DP3', 2), ('PGM5P2', 2), ('UBE2MP1', 2), ('HAR1\n",
      "169 input query terms found no hit:\t['LOC284551', 'LOC285548', 'LOC728410', 'LOC541473', 'DULLARD', 'KIAA0368', 'EFTUD1', 'TWISTNB', 'SF\n",
      "13 input query terms found dup hits:\t[('S100A7L2', 2), ('POM121L8P', 2), ('MEG8', 2), ('KIR3DX1', 5), ('RFPL1S', 2), ('SNORD91B', 2), ('C\n",
      "165 input query terms found no hit:\t['TMEM188', 'PDZD3', 'FAM102B', 'FAM102A', 'SMCR7L', 'G6PC', 'OSTCL', 'LOC653544', 'LOC653545', 'USP\n",
      "16 input query terms found dup hits:\t[('PCNAP1', 2), ('SNORA63', 6), ('SERHL', 2), ('CEACAM22P', 2), ('SNORA16A', 2), ('FAM41AY1', 2), ('\n",
      "147 input query terms found no hit:\t['LRRC37A4', 'LOC100131726', 'CPSF3L', 'COL4A3BP', 'PAR1', 'LOC92973', 'MICALCL', 'SMCR7', 'HIST4H4'\n",
      "15 input query terms found dup hits:\t[('MBL1P', 2), ('SDHAP3', 2), ('PSORS1C3', 8), ('MYADML', 2), ('POM121L10P', 2), ('HLA-J', 9), ('HLA\n",
      "153 input query terms found no hit:\t['LOC100131193', 'KIAA1109', 'DUS2L', 'MYST1', 'IGLL3', 'C12orf69', 'NEURL', 'FAM188B', 'FLJ32063', \n",
      "9 input query terms found dup hits:\t[('PLGLA', 2), ('PA2G4P4', 2), ('RPL23AP7', 2), ('RPL23AP64', 3), ('RPSAP52', 2), ('TPRXL', 2), ('DP\n",
      "177 input query terms found no hit:\t['LOC283314', 'RBM9', 'C20orf118', 'LOC150786', 'C20orf112', 'C20orf111', 'C20orf117', 'C20orf114', \n",
      "18 input query terms found dup hits:\t[('TPTE2P1', 2), ('FOLH1B', 2), ('ID2B', 2), ('MSX2P1', 2), ('RAET1K', 2), ('BMS1P4', 2), ('RPL23AP8\n",
      "175 input query terms found no hit:\t['MYEOV2', 'FAM35B2', 'C20orf85', 'FAM175A', 'C17orf73', 'C17orf72', 'C17orf71', 'C17orf70', 'C17orf\n",
      "19 input query terms found dup hits:\t[('CATSPER2P1', 2), ('AQP7P1', 2), ('ATP8B5P', 2), ('OR7E91P', 2), ('ZNF204P', 2), ('CECR7', 2), ('M\n",
      "144 input query terms found no hit:\t['BASE', 'C14orf21', 'C14orf23', 'FAM5B', 'TROVE2', 'C18orf62', 'LOC401052', 'GNB2L1', 'FAM19A4', 'F\n",
      "13 input query terms found dup hits:\t[('P2RX6P', 2), ('RP9P', 2), ('NRADDP', 2), ('MORF4', 2), ('SNORA75', 6), ('SNORA72', 2), ('FMO6P', \n",
      "168 input query terms found no hit:\t['TTC15', 'C4orf29', 'FAM48A', 'C4orf21', 'C3orf65', 'C4orf23', 'C3orf67', 'LOC144742', 'C3orf63', '\n",
      "19 input query terms found dup hits:\t[('RPL13AP17', 2), ('DNAJB3', 2), ('SNORA58', 2), ('PMCHL1', 2), ('FAM95B1', 2), ('TMEM191A', 2), ('\n",
      "144 input query terms found no hit:\t['LOC254312', 'ZNF542', 'CSDAP1', 'FAM73A', 'FAM73B', 'TCTE3', 'ZNF37B', 'LOC728875', 'KIAA1826', 'P\n",
      "15 input query terms found dup hits:\t[('ZFP91-CNTF', 2), ('PRSS30P', 3), ('GSTM2P1', 2), ('C3P1', 2), ('NCF1B', 2), ('SNORD114-4', 2), ('\n",
      "138 input query terms found no hit:\t['LOC100129716', 'C14orf184', 'C14orf181', 'C14orf183', 'C14orf182', 'C7orf26', 'C7orf27', 'C7orf23'\n",
      "14 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P3', 2), ('ANXA2P2', 2), ('ZNF286B', 2), ('SAA3P', 2), ('GEMIN8P4', 2), ('DI\n",
      "132 input query terms found no hit:\t['AKR1CL1', 'CCDC21', 'CCDC23', 'TMEM111', 'APOB48R', 'LOC595101', 'KIAA1430', 'KIAA1432', 'ORC4L', \n",
      "11 input query terms found dup hits:\t[('TCP10L2', 2), ('FTLP10', 2), ('TPTE2P3', 2), ('VENTXP1', 2), ('XKRY', 2), ('HTR7P1', 2), ('STAG3L\n",
      "175 input query terms found no hit:\t['DFNB59', 'LOC100170939', 'FAM36A', 'C2orf70', 'C2orf71', 'C2orf77', 'C2orf79', 'LOC100128191', 'HR\n",
      "16 input query terms found dup hits:\t[('ZNF702P', 2), ('BRD7P3', 2), ('SCARNA17', 2), ('GABARAPL3', 2), ('MT1IP', 2), ('ADAM6', 3), ('PDZ\n",
      "142 input query terms found no hit:\t['C9orf30', 'PEG3AS', 'C9orf37', 'TXNDC3', 'LOC442308', 'FAM46D', 'FAM46A', 'FAM46B', 'FAM46C', 'WDR\n",
      "10 input query terms found dup hits:\t[('FAM197Y2', 3), ('SNORA71B', 2), ('PRKY', 2), ('NAT8B', 3), ('VENTXP7', 2), ('SPRR2C', 2), ('RPL23\n",
      "148 input query terms found no hit:\t['FCGR1B', 'RG9MTD2', 'RARS', 'LOC727677', 'CCRN4L', 'SEPT14', 'HKR1', 'C1orf112', 'LOC100270710', '\n",
      "16 input query terms found dup hits:\t[('OR6W1P', 3), ('LYPLA2P1', 6), ('RPL23AP53', 2), ('TRPC2', 2), ('RBMY2FP', 2), ('BAGE2', 2), ('POM\n",
      "150 input query terms found no hit:\t['SIP1', 'FAM129C', 'GIF', 'TMCO7', 'CXorf61', 'CXorf64', 'ZNF192', 'SEPP1', 'KIRREL', 'MGC16121', '\n",
      "17 input query terms found dup hits:\t[('ISCA1P1', 2), ('PCDHB19P', 3), ('NAPSB', 2), ('WASH7P', 2), ('TTTY11', 2), ('HCG4', 9), ('OR5E1P'\n",
      "135 input query terms found no hit:\t['PPYR1', 'FAM82A2', 'FLJ37453', 'FAM82A1', 'KIAA0947', 'MOBKL2B', 'MOBKL2C', 'MOBKL2A', 'CMAH', 'GR\n",
      "5 input query terms found dup hits:\t[('GGNBP1', 2), ('WASH5P', 2), ('RPL29P2', 2), ('CDR1', 2), ('OR10J3', 3)]\n",
      "78 input query terms found no hit:\t['C14orf174', 'C14orf176', 'C14orf177', 'C14orf179', 'PTPLB', 'SGEF', 'LOC400752', 'GRLF1', 'SPANXE'\n"
     ]
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.108889749Z",
     "start_time": "2024-01-10T21:38:18.021199718Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J1-01</th>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.641092</td>\n",
       "      <td>-0.325826</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.266428</td>\n",
       "      <td>0.355422</td>\n",
       "      <td>0.03719</td>\n",
       "      <td>0.243706</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.520186</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.182978</td>\n",
       "      <td>-0.615817</td>\n",
       "      <td>-0.281533</td>\n",
       "      <td>3.02111</td>\n",
       "      <td>-0.927577</td>\n",
       "      <td>-1.006227</td>\n",
       "      <td>1.119905</td>\n",
       "      <td>-2.185533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J2-01</th>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.864792</td>\n",
       "      <td>2.766674</td>\n",
       "      <td>0.321165</td>\n",
       "      <td>1.000728</td>\n",
       "      <td>0.836122</td>\n",
       "      <td>0.35439</td>\n",
       "      <td>-0.436694</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.318586</td>\n",
       "      <td>1.056018</td>\n",
       "      <td>0.393822</td>\n",
       "      <td>2.366583</td>\n",
       "      <td>-0.955033</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>1.020723</td>\n",
       "      <td>1.226373</td>\n",
       "      <td>1.164005</td>\n",
       "      <td>0.265067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J3-01</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.723192</td>\n",
       "      <td>-0.362926</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.639828</td>\n",
       "      <td>-0.199578</td>\n",
       "      <td>-0.48331</td>\n",
       "      <td>0.143606</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.574486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>-0.113317</td>\n",
       "      <td>-3.803333</td>\n",
       "      <td>-0.61009</td>\n",
       "      <td>0.397623</td>\n",
       "      <td>-0.675227</td>\n",
       "      <td>1.196005</td>\n",
       "      <td>-3.161633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J5-01</th>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.576792</td>\n",
       "      <td>-2.086226</td>\n",
       "      <td>2.463765</td>\n",
       "      <td>1.382228</td>\n",
       "      <td>-1.115678</td>\n",
       "      <td>-1.23621</td>\n",
       "      <td>0.615806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.279486</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.078622</td>\n",
       "      <td>1.095983</td>\n",
       "      <td>-0.908533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>0.661823</td>\n",
       "      <td>0.458273</td>\n",
       "      <td>0.839605</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-OR-A5J6-01</th>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.311992</td>\n",
       "      <td>5.225974</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.967928</td>\n",
       "      <td>-0.393778</td>\n",
       "      <td>-0.38231</td>\n",
       "      <td>-0.060194</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>1.607218</td>\n",
       "      <td>2.481122</td>\n",
       "      <td>-0.946617</td>\n",
       "      <td>-0.570533</td>\n",
       "      <td>-1.28139</td>\n",
       "      <td>-0.425177</td>\n",
       "      <td>0.938573</td>\n",
       "      <td>0.495005</td>\n",
       "      <td>-1.733333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Anxiety disorder  Age  Gender  ARHGEF10L     HIF3A     RNF17  \\\n",
       "sampleID                                                                        \n",
       "TCGA-OR-A5J1-01                 1   58       1  -0.641092 -0.325826 -0.531035   \n",
       "TCGA-OR-A5J2-01                 1   44       0  -1.864792  2.766674  0.321165   \n",
       "TCGA-OR-A5J3-01                 1   23       0  -0.723192 -0.362926 -0.531035   \n",
       "TCGA-OR-A5J5-01                 1   30       1  -1.576792 -2.086226  2.463765   \n",
       "TCGA-OR-A5J6-01                 1   29       0  -2.311992  5.225974 -0.531035   \n",
       "\n",
       "                    RNF10     RNF11    RNF13   GTF2IP1  ...   SLC7A10  \\\n",
       "sampleID                                                ...             \n",
       "TCGA-OR-A5J1-01  1.266428  0.355422  0.03719  0.243706  ... -1.520186   \n",
       "TCGA-OR-A5J2-01  1.000728  0.836122  0.35439 -0.436694  ... -0.318586   \n",
       "TCGA-OR-A5J3-01  0.639828 -0.199578 -0.48331  0.143606  ... -0.574486   \n",
       "TCGA-OR-A5J5-01  1.382228 -1.115678 -1.23621  0.615806  ... -0.279486   \n",
       "TCGA-OR-A5J6-01  0.967928 -0.393778 -0.38231 -0.060194  ... -2.090786   \n",
       "\n",
       "                  PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1     TULP3  \\\n",
       "sampleID                                                                     \n",
       "TCGA-OR-A5J1-01 -0.086682 -0.182978 -0.615817 -0.281533  3.02111 -0.927577   \n",
       "TCGA-OR-A5J2-01  1.056018  0.393822  2.366583 -0.955033 -1.28139  1.020723   \n",
       "TCGA-OR-A5J3-01 -0.086682 -0.748878 -0.113317 -3.803333 -0.61009  0.397623   \n",
       "TCGA-OR-A5J5-01 -0.086682  0.078622  1.095983 -0.908533 -1.28139  0.661823   \n",
       "TCGA-OR-A5J6-01  1.607218  2.481122 -0.946617 -0.570533 -1.28139 -0.425177   \n",
       "\n",
       "                    BCL6B     GSTK1      SELP  \n",
       "sampleID                                       \n",
       "TCGA-OR-A5J1-01 -1.006227  1.119905 -2.185533  \n",
       "TCGA-OR-A5J2-01  1.226373  1.164005  0.265067  \n",
       "TCGA-OR-A5J3-01 -0.675227  1.196005 -3.161633  \n",
       "TCGA-OR-A5J5-01  0.458273  0.839605 -5.525533  \n",
       "TCGA-OR-A5J6-01  0.938573  0.495005 -1.733333  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.112793384Z",
     "start_time": "2024-01-10T21:38:18.107415526Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 79 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1' with 79 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 35.0\n",
      "  50% (Median): 49.0\n",
      "  75%: 59.5\n",
      "Min: 14\n",
      "Max: 77\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1' with 31 occurrences. This represents 39.24% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.377898533Z",
     "start_time": "2024-01-10T21:38:18.111417452Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425336629Z",
     "start_time": "2024-01-10T21:38:31.419043732Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output\\Jiayi\\Anxiety-disorder\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE119995',\n",
       " 'GSE18123',\n",
       " 'GSE60190',\n",
       " 'GSE60491',\n",
       " 'GSE61672',\n",
       " 'GSE68526',\n",
       " 'GSE78104',\n",
       " 'GSE94119',\n",
       " 'GSE98793']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Anxiety-disorder\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:02.935079426Z",
     "start_time": "2024-01-10T21:39:02.923698385Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Genetic Neuropathology of Obsessive Psychiatric Syndromes\"\n",
      "!Series_summary\t\"Anorexia nervosa (AN), bulimia nervosa (BN), and obsessive-compulsive disorder (OCD) are complex psychiatric disorders with shared obsessive features, thought to arise from the interaction of multiple genes of small effect with environmental factors.  Potential candidate genes for AN, BN, and OCD have been identified through clinical association and neuroimaging studies; however, recent genome-wide association studies of eating disorders (ED) so far have failed to report significant findings. Additionally, few if any studies have interrogated postmortem brain tissue for evidence of eQTLs associated with candidate genes, which has particular promise as an approach to elucidating molecular mechanisms of association. We therefore selected single nucleotide polymorphisms (SNPs) based on candidate gene studies for AN, BN, and OCD from the literature, and examined the association of these SNPs with gene expression across the lifespan in prefrontal cortex of a non-psychiatric control cohort (N=268).   Several risk-predisposing SNPs were significantly associated with gene expression among control subjects. We then measured gene expression in the prefrontal cortex of cases previously diagnosed with obsessive psychiatric disorders, e.g., eating disorders (ED; N=15), and obsessive-compulsive disorder/obsessive-compulsive personality disorder or tics (OCD/OCPD/Tic; N=16), and non-psychiatric controls (N=102) and identified 6 and 286 genes that were differentially expressed between ED compared to controls and OCD cases compared to controls, respectively (FDR < 5%). However, none of the clinical risk SNPs were among the eQTLs and none were significantly associated with gene expression within the broad obsessive cohort, suggesting larger sample sizes or other brain regions may be required to identify candidate molecular mechanisms of clinical association in postmortem brain datasets.\"\n",
      "!Series_overall_design\t\"Gene expression data from the dorsolateral prefrontal cortex (DLPFC) from postmortem tissue on 133 subjects - 15 eating disorder (ED) patients, 16 obessive compulsive disorder (OCD) patients, and 102 non-psychiatric controls - run on the Illumina HumanHT-12 v3 microarray\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1467273</th>\n",
       "      <th>GSM1467274</th>\n",
       "      <th>GSM1467275</th>\n",
       "      <th>GSM1467276</th>\n",
       "      <th>GSM1467277</th>\n",
       "      <th>GSM1467278</th>\n",
       "      <th>GSM1467279</th>\n",
       "      <th>GSM1467280</th>\n",
       "      <th>GSM1467281</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1467396</th>\n",
       "      <th>GSM1467397</th>\n",
       "      <th>GSM1467398</th>\n",
       "      <th>GSM1467399</th>\n",
       "      <th>GSM1467400</th>\n",
       "      <th>GSM1467401</th>\n",
       "      <th>GSM1467402</th>\n",
       "      <th>GSM1467403</th>\n",
       "      <th>GSM1467404</th>\n",
       "      <th>GSM1467405</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>rin: 7.4</td>\n",
       "      <td>rin: 8.6</td>\n",
       "      <td>rin: 7.8</td>\n",
       "      <td>rin: 8.2</td>\n",
       "      <td>rin: 8.5</td>\n",
       "      <td>rin: 8.3</td>\n",
       "      <td>rin: 8.1</td>\n",
       "      <td>rin: 8.8</td>\n",
       "      <td>rin: 8.7</td>\n",
       "      <td>...</td>\n",
       "      <td>rin: 8.4</td>\n",
       "      <td>rin: 8</td>\n",
       "      <td>rin: 8.3</td>\n",
       "      <td>rin: 7.5</td>\n",
       "      <td>rin: 8.1</td>\n",
       "      <td>rin: 8.5</td>\n",
       "      <td>rin: 6.6</td>\n",
       "      <td>rin: 7.1</td>\n",
       "      <td>rin: 9</td>\n",
       "      <td>rin: 8.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>ocd: ED</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: OCD</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: ED</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>...</td>\n",
       "      <td>ocd: ED</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: ED</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "      <td>ocd: Control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 1</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "      <td>rinmatched: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>dx: Bipolar</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Bipolar</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Bipolar</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>...</td>\n",
       "      <td>dx: ED</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: MDD</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "      <td>dx: Control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>ph: 6.18</td>\n",
       "      <td>ph: 6.59</td>\n",
       "      <td>ph: 6.37</td>\n",
       "      <td>ph: 6.6</td>\n",
       "      <td>ph: 6.38</td>\n",
       "      <td>ph: 6.02</td>\n",
       "      <td>ph: 6.87</td>\n",
       "      <td>ph: 6.95</td>\n",
       "      <td>ph: 6.82</td>\n",
       "      <td>...</td>\n",
       "      <td>ph: 6.34</td>\n",
       "      <td>ph: 6.5</td>\n",
       "      <td>ph: 6.51</td>\n",
       "      <td>ph: 6.59</td>\n",
       "      <td>ph: 6.65</td>\n",
       "      <td>ph: 6.71</td>\n",
       "      <td>ph: 6.2</td>\n",
       "      <td>ph: 6.2</td>\n",
       "      <td>ph: 6.83</td>\n",
       "      <td>ph: 6.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 50.421917</td>\n",
       "      <td>age: 27.49863</td>\n",
       "      <td>age: 30.627397</td>\n",
       "      <td>age: 61.167123</td>\n",
       "      <td>age: 32.69589</td>\n",
       "      <td>age: 39.213698</td>\n",
       "      <td>age: 58.605479</td>\n",
       "      <td>age: 49.2</td>\n",
       "      <td>age: 41.041095</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 30.726027</td>\n",
       "      <td>age: 63.471232</td>\n",
       "      <td>age: 54.808219</td>\n",
       "      <td>age: 57.512328</td>\n",
       "      <td>age: 57.610958</td>\n",
       "      <td>age: 44.958904</td>\n",
       "      <td>age: 35.684931</td>\n",
       "      <td>age: 63</td>\n",
       "      <td>age: 38.780821</td>\n",
       "      <td>age: 45.978082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>pmi: 27</td>\n",
       "      <td>pmi: 19.5</td>\n",
       "      <td>pmi: 71.5</td>\n",
       "      <td>pmi: 19.5</td>\n",
       "      <td>pmi: 22.5</td>\n",
       "      <td>pmi: 22.5</td>\n",
       "      <td>pmi: 64</td>\n",
       "      <td>pmi: 28</td>\n",
       "      <td>pmi: 18</td>\n",
       "      <td>...</td>\n",
       "      <td>pmi: 19</td>\n",
       "      <td>pmi: 37</td>\n",
       "      <td>pmi: 21</td>\n",
       "      <td>pmi: 24</td>\n",
       "      <td>pmi: 17</td>\n",
       "      <td>pmi: 20</td>\n",
       "      <td>pmi: 33</td>\n",
       "      <td>pmi: 45.5</td>\n",
       "      <td>pmi: 29</td>\n",
       "      <td>pmi: 34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>...</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "      <td>race: CAUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>batch1: 16</td>\n",
       "      <td>batch1: 18</td>\n",
       "      <td>batch1: 18</td>\n",
       "      <td>batch1: 18</td>\n",
       "      <td>batch1: 19</td>\n",
       "      <td>batch1: 19</td>\n",
       "      <td>batch1: 20</td>\n",
       "      <td>batch1: 21</td>\n",
       "      <td>batch1: 21</td>\n",
       "      <td>...</td>\n",
       "      <td>batch1: 96</td>\n",
       "      <td>batch1: 96</td>\n",
       "      <td>batch1: 97</td>\n",
       "      <td>batch1: 98</td>\n",
       "      <td>batch1: 99</td>\n",
       "      <td>batch1: 99</td>\n",
       "      <td>batch1: 100</td>\n",
       "      <td>batch1: 102</td>\n",
       "      <td>batch1: 102</td>\n",
       "      <td>batch1: 102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 134 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession      GSM1467273     GSM1467274      GSM1467275  \\\n",
       "0  !Sample_characteristics_ch1        rin: 7.4       rin: 8.6        rin: 7.8   \n",
       "1  !Sample_characteristics_ch1         ocd: ED   ocd: Control        ocd: OCD   \n",
       "2  !Sample_characteristics_ch1   rinmatched: 1  rinmatched: 0   rinmatched: 1   \n",
       "3  !Sample_characteristics_ch1     dx: Bipolar    dx: Control     dx: Bipolar   \n",
       "4  !Sample_characteristics_ch1        ph: 6.18       ph: 6.59        ph: 6.37   \n",
       "5  !Sample_characteristics_ch1  age: 50.421917  age: 27.49863  age: 30.627397   \n",
       "6  !Sample_characteristics_ch1         pmi: 27      pmi: 19.5       pmi: 71.5   \n",
       "7  !Sample_characteristics_ch1          Sex: F         Sex: M          Sex: M   \n",
       "8  !Sample_characteristics_ch1      race: CAUC     race: CAUC      race: CAUC   \n",
       "9  !Sample_characteristics_ch1      batch1: 16     batch1: 18      batch1: 18   \n",
       "\n",
       "       GSM1467276     GSM1467277      GSM1467278      GSM1467279  \\\n",
       "0        rin: 8.2       rin: 8.5        rin: 8.3        rin: 8.1   \n",
       "1    ocd: Control   ocd: Control         ocd: ED    ocd: Control   \n",
       "2   rinmatched: 0  rinmatched: 0   rinmatched: 1   rinmatched: 0   \n",
       "3     dx: Control    dx: Control     dx: Bipolar     dx: Control   \n",
       "4         ph: 6.6       ph: 6.38        ph: 6.02        ph: 6.87   \n",
       "5  age: 61.167123  age: 32.69589  age: 39.213698  age: 58.605479   \n",
       "6       pmi: 19.5      pmi: 22.5       pmi: 22.5         pmi: 64   \n",
       "7          Sex: M         Sex: M          Sex: F          Sex: M   \n",
       "8      race: CAUC     race: CAUC      race: CAUC      race: CAUC   \n",
       "9      batch1: 18     batch1: 19      batch1: 19      batch1: 20   \n",
       "\n",
       "      GSM1467280      GSM1467281  ...      GSM1467396      GSM1467397  \\\n",
       "0       rin: 8.8        rin: 8.7  ...        rin: 8.4          rin: 8   \n",
       "1   ocd: Control    ocd: Control  ...         ocd: ED    ocd: Control   \n",
       "2  rinmatched: 0   rinmatched: 0  ...   rinmatched: 1   rinmatched: 1   \n",
       "3    dx: Control     dx: Control  ...          dx: ED     dx: Control   \n",
       "4       ph: 6.95        ph: 6.82  ...        ph: 6.34         ph: 6.5   \n",
       "5      age: 49.2  age: 41.041095  ...  age: 30.726027  age: 63.471232   \n",
       "6        pmi: 28         pmi: 18  ...         pmi: 19         pmi: 37   \n",
       "7         Sex: M          Sex: M  ...          Sex: F          Sex: M   \n",
       "8     race: CAUC      race: CAUC  ...      race: CAUC      race: CAUC   \n",
       "9     batch1: 21      batch1: 21  ...      batch1: 96      batch1: 96   \n",
       "\n",
       "       GSM1467398      GSM1467399      GSM1467400      GSM1467401  \\\n",
       "0        rin: 8.3        rin: 7.5        rin: 8.1        rin: 8.5   \n",
       "1    ocd: Control    ocd: Control    ocd: Control    ocd: Control   \n",
       "2   rinmatched: 0   rinmatched: 1   rinmatched: 1   rinmatched: 0   \n",
       "3     dx: Control     dx: Control     dx: Control     dx: Control   \n",
       "4        ph: 6.51        ph: 6.59        ph: 6.65        ph: 6.71   \n",
       "5  age: 54.808219  age: 57.512328  age: 57.610958  age: 44.958904   \n",
       "6         pmi: 21         pmi: 24         pmi: 17         pmi: 20   \n",
       "7          Sex: M          Sex: M          Sex: M          Sex: M   \n",
       "8      race: CAUC      race: CAUC      race: CAUC      race: CAUC   \n",
       "9      batch1: 97      batch1: 98      batch1: 99      batch1: 99   \n",
       "\n",
       "       GSM1467402     GSM1467403      GSM1467404      GSM1467405  \n",
       "0        rin: 6.6       rin: 7.1          rin: 9        rin: 8.3  \n",
       "1         ocd: ED   ocd: Control    ocd: Control    ocd: Control  \n",
       "2   rinmatched: 1  rinmatched: 1   rinmatched: 0   rinmatched: 0  \n",
       "3         dx: MDD    dx: Control     dx: Control     dx: Control  \n",
       "4         ph: 6.2        ph: 6.2        ph: 6.83        ph: 6.92  \n",
       "5  age: 35.684931        age: 63  age: 38.780821  age: 45.978082  \n",
       "6         pmi: 33      pmi: 45.5         pmi: 29         pmi: 34  \n",
       "7          Sex: F         Sex: M          Sex: M          Sex: M  \n",
       "8      race: CAUC     race: CAUC      race: CAUC      race: CAUC  \n",
       "9     batch1: 100    batch1: 102     batch1: 102     batch1: 102  \n",
       "\n",
       "[10 rows x 134 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE60190\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'ocd: ED', 'ocd: Control',\n",
       "       'ocd: OCD'], dtype=object)"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 5\n",
    "gender_row = 7\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'ocd: OCD':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "    \n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = round(float(age_string.split(': ')[1]),2)\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1467273</th>\n",
       "      <th>GSM1467274</th>\n",
       "      <th>GSM1467275</th>\n",
       "      <th>GSM1467276</th>\n",
       "      <th>GSM1467277</th>\n",
       "      <th>GSM1467278</th>\n",
       "      <th>GSM1467279</th>\n",
       "      <th>GSM1467280</th>\n",
       "      <th>GSM1467281</th>\n",
       "      <th>GSM1467282</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1467396</th>\n",
       "      <th>GSM1467397</th>\n",
       "      <th>GSM1467398</th>\n",
       "      <th>GSM1467399</th>\n",
       "      <th>GSM1467400</th>\n",
       "      <th>GSM1467401</th>\n",
       "      <th>GSM1467402</th>\n",
       "      <th>GSM1467403</th>\n",
       "      <th>GSM1467404</th>\n",
       "      <th>GSM1467405</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>50.42</td>\n",
       "      <td>27.5</td>\n",
       "      <td>30.63</td>\n",
       "      <td>61.17</td>\n",
       "      <td>32.7</td>\n",
       "      <td>39.21</td>\n",
       "      <td>58.61</td>\n",
       "      <td>49.2</td>\n",
       "      <td>41.04</td>\n",
       "      <td>51.75</td>\n",
       "      <td>...</td>\n",
       "      <td>30.73</td>\n",
       "      <td>63.47</td>\n",
       "      <td>54.81</td>\n",
       "      <td>57.51</td>\n",
       "      <td>57.61</td>\n",
       "      <td>44.96</td>\n",
       "      <td>35.68</td>\n",
       "      <td>63.0</td>\n",
       "      <td>38.78</td>\n",
       "      <td>45.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM1467273  GSM1467274  GSM1467275  GSM1467276  GSM1467277  \\\n",
       "Anxiety disorder        0.00         0.0        1.00        0.00         0.0   \n",
       "Age                    50.42        27.5       30.63       61.17        32.7   \n",
       "Gender                  1.00         0.0        0.00        0.00         0.0   \n",
       "\n",
       "                  GSM1467278  GSM1467279  GSM1467280  GSM1467281  GSM1467282  \\\n",
       "Anxiety disorder        0.00        0.00         0.0        0.00        0.00   \n",
       "Age                    39.21       58.61        49.2       41.04       51.75   \n",
       "Gender                  1.00        0.00         0.0        0.00        0.00   \n",
       "\n",
       "                  ...  GSM1467396  GSM1467397  GSM1467398  GSM1467399  \\\n",
       "Anxiety disorder  ...        0.00        0.00        0.00        0.00   \n",
       "Age               ...       30.73       63.47       54.81       57.51   \n",
       "Gender            ...        1.00        0.00        0.00        0.00   \n",
       "\n",
       "                  GSM1467400  GSM1467401  GSM1467402  GSM1467403  GSM1467404  \\\n",
       "Anxiety disorder        0.00        0.00        0.00         0.0        0.00   \n",
       "Age                    57.61       44.96       35.68        63.0       38.78   \n",
       "Gender                  0.00        0.00        1.00         0.0        0.00   \n",
       "\n",
       "                  GSM1467405  \n",
       "Anxiety disorder        0.00  \n",
       "Age                    45.98  \n",
       "Gender                  0.00  \n",
       "\n",
       "[3 rows x 133 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1467273</th>\n",
       "      <th>GSM1467274</th>\n",
       "      <th>GSM1467275</th>\n",
       "      <th>GSM1467276</th>\n",
       "      <th>GSM1467277</th>\n",
       "      <th>GSM1467278</th>\n",
       "      <th>GSM1467279</th>\n",
       "      <th>GSM1467280</th>\n",
       "      <th>GSM1467281</th>\n",
       "      <th>GSM1467282</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1467396</th>\n",
       "      <th>GSM1467397</th>\n",
       "      <th>GSM1467398</th>\n",
       "      <th>GSM1467399</th>\n",
       "      <th>GSM1467400</th>\n",
       "      <th>GSM1467401</th>\n",
       "      <th>GSM1467402</th>\n",
       "      <th>GSM1467403</th>\n",
       "      <th>GSM1467404</th>\n",
       "      <th>GSM1467405</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>14.631928</td>\n",
       "      <td>14.846268</td>\n",
       "      <td>14.949614</td>\n",
       "      <td>15.044821</td>\n",
       "      <td>14.790150</td>\n",
       "      <td>14.670395</td>\n",
       "      <td>14.730323</td>\n",
       "      <td>14.972534</td>\n",
       "      <td>14.826768</td>\n",
       "      <td>14.790150</td>\n",
       "      <td>...</td>\n",
       "      <td>14.749395</td>\n",
       "      <td>14.593207</td>\n",
       "      <td>14.558687</td>\n",
       "      <td>14.490661</td>\n",
       "      <td>14.540732</td>\n",
       "      <td>14.540732</td>\n",
       "      <td>14.972534</td>\n",
       "      <td>14.730323</td>\n",
       "      <td>14.790150</td>\n",
       "      <td>14.609839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1343295</th>\n",
       "      <td>13.015539</td>\n",
       "      <td>13.199489</td>\n",
       "      <td>13.256768</td>\n",
       "      <td>13.030246</td>\n",
       "      <td>12.611440</td>\n",
       "      <td>13.030246</td>\n",
       "      <td>12.894845</td>\n",
       "      <td>12.668329</td>\n",
       "      <td>12.536053</td>\n",
       "      <td>12.684084</td>\n",
       "      <td>...</td>\n",
       "      <td>12.729559</td>\n",
       "      <td>12.628114</td>\n",
       "      <td>12.729559</td>\n",
       "      <td>12.526797</td>\n",
       "      <td>12.921417</td>\n",
       "      <td>12.536053</td>\n",
       "      <td>12.616903</td>\n",
       "      <td>12.685365</td>\n",
       "      <td>12.729559</td>\n",
       "      <td>12.894845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651199</th>\n",
       "      <td>7.407451</td>\n",
       "      <td>7.404480</td>\n",
       "      <td>7.384722</td>\n",
       "      <td>7.398199</td>\n",
       "      <td>7.388903</td>\n",
       "      <td>7.392256</td>\n",
       "      <td>7.411295</td>\n",
       "      <td>7.392441</td>\n",
       "      <td>7.402658</td>\n",
       "      <td>7.397018</td>\n",
       "      <td>...</td>\n",
       "      <td>7.386564</td>\n",
       "      <td>7.392892</td>\n",
       "      <td>7.387466</td>\n",
       "      <td>7.398703</td>\n",
       "      <td>7.387851</td>\n",
       "      <td>7.380980</td>\n",
       "      <td>7.391920</td>\n",
       "      <td>7.391737</td>\n",
       "      <td>7.384270</td>\n",
       "      <td>7.391287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651209</th>\n",
       "      <td>7.551069</td>\n",
       "      <td>7.489157</td>\n",
       "      <td>7.481180</td>\n",
       "      <td>7.454187</td>\n",
       "      <td>7.468150</td>\n",
       "      <td>7.538856</td>\n",
       "      <td>7.516444</td>\n",
       "      <td>7.428403</td>\n",
       "      <td>7.420693</td>\n",
       "      <td>7.435272</td>\n",
       "      <td>...</td>\n",
       "      <td>7.492618</td>\n",
       "      <td>7.517797</td>\n",
       "      <td>7.570776</td>\n",
       "      <td>7.491987</td>\n",
       "      <td>7.543178</td>\n",
       "      <td>7.536219</td>\n",
       "      <td>7.486087</td>\n",
       "      <td>7.445162</td>\n",
       "      <td>7.523485</td>\n",
       "      <td>7.524602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651210</th>\n",
       "      <td>7.389822</td>\n",
       "      <td>7.396480</td>\n",
       "      <td>7.392958</td>\n",
       "      <td>7.389106</td>\n",
       "      <td>7.399044</td>\n",
       "      <td>7.401760</td>\n",
       "      <td>7.412553</td>\n",
       "      <td>7.418004</td>\n",
       "      <td>7.383820</td>\n",
       "      <td>7.399614</td>\n",
       "      <td>...</td>\n",
       "      <td>7.383362</td>\n",
       "      <td>7.386934</td>\n",
       "      <td>7.383795</td>\n",
       "      <td>7.390598</td>\n",
       "      <td>7.383517</td>\n",
       "      <td>7.380075</td>\n",
       "      <td>7.383755</td>\n",
       "      <td>7.392136</td>\n",
       "      <td>7.383855</td>\n",
       "      <td>7.395560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415911</th>\n",
       "      <td>8.200796</td>\n",
       "      <td>8.057956</td>\n",
       "      <td>8.244277</td>\n",
       "      <td>8.318973</td>\n",
       "      <td>8.585545</td>\n",
       "      <td>8.051581</td>\n",
       "      <td>8.212767</td>\n",
       "      <td>8.391308</td>\n",
       "      <td>8.099802</td>\n",
       "      <td>8.173629</td>\n",
       "      <td>...</td>\n",
       "      <td>8.142346</td>\n",
       "      <td>8.187782</td>\n",
       "      <td>8.310146</td>\n",
       "      <td>8.234435</td>\n",
       "      <td>8.165085</td>\n",
       "      <td>8.222113</td>\n",
       "      <td>7.763713</td>\n",
       "      <td>8.091295</td>\n",
       "      <td>8.353276</td>\n",
       "      <td>7.976434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415926</th>\n",
       "      <td>9.308517</td>\n",
       "      <td>9.261562</td>\n",
       "      <td>9.539295</td>\n",
       "      <td>8.953072</td>\n",
       "      <td>9.178453</td>\n",
       "      <td>9.364314</td>\n",
       "      <td>9.391644</td>\n",
       "      <td>9.428445</td>\n",
       "      <td>9.813872</td>\n",
       "      <td>9.575225</td>\n",
       "      <td>...</td>\n",
       "      <td>9.629529</td>\n",
       "      <td>9.005815</td>\n",
       "      <td>9.073252</td>\n",
       "      <td>9.540430</td>\n",
       "      <td>9.374462</td>\n",
       "      <td>9.614333</td>\n",
       "      <td>8.026149</td>\n",
       "      <td>8.558921</td>\n",
       "      <td>8.787474</td>\n",
       "      <td>9.170768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415949</th>\n",
       "      <td>8.258738</td>\n",
       "      <td>8.050959</td>\n",
       "      <td>8.363362</td>\n",
       "      <td>8.037019</td>\n",
       "      <td>8.059074</td>\n",
       "      <td>8.154139</td>\n",
       "      <td>8.092218</td>\n",
       "      <td>8.098232</td>\n",
       "      <td>8.231403</td>\n",
       "      <td>8.278169</td>\n",
       "      <td>...</td>\n",
       "      <td>8.041222</td>\n",
       "      <td>7.976685</td>\n",
       "      <td>7.960897</td>\n",
       "      <td>7.998456</td>\n",
       "      <td>8.063397</td>\n",
       "      <td>7.888010</td>\n",
       "      <td>7.992915</td>\n",
       "      <td>8.265580</td>\n",
       "      <td>7.960649</td>\n",
       "      <td>8.072490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415979</th>\n",
       "      <td>11.414900</td>\n",
       "      <td>9.318214</td>\n",
       "      <td>9.597813</td>\n",
       "      <td>9.561681</td>\n",
       "      <td>9.837778</td>\n",
       "      <td>9.649293</td>\n",
       "      <td>10.331787</td>\n",
       "      <td>10.224846</td>\n",
       "      <td>10.343168</td>\n",
       "      <td>10.534634</td>\n",
       "      <td>...</td>\n",
       "      <td>10.500506</td>\n",
       "      <td>10.723755</td>\n",
       "      <td>10.533912</td>\n",
       "      <td>10.080191</td>\n",
       "      <td>11.034512</td>\n",
       "      <td>10.533200</td>\n",
       "      <td>11.493401</td>\n",
       "      <td>10.312506</td>\n",
       "      <td>10.355448</td>\n",
       "      <td>10.214869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2416019</th>\n",
       "      <td>7.386126</td>\n",
       "      <td>7.390666</td>\n",
       "      <td>7.386612</td>\n",
       "      <td>7.385291</td>\n",
       "      <td>7.383622</td>\n",
       "      <td>7.387200</td>\n",
       "      <td>7.403395</td>\n",
       "      <td>7.388422</td>\n",
       "      <td>7.381764</td>\n",
       "      <td>7.389056</td>\n",
       "      <td>...</td>\n",
       "      <td>7.386483</td>\n",
       "      <td>7.417538</td>\n",
       "      <td>7.380281</td>\n",
       "      <td>7.383249</td>\n",
       "      <td>7.375744</td>\n",
       "      <td>7.383861</td>\n",
       "      <td>7.384510</td>\n",
       "      <td>7.393648</td>\n",
       "      <td>7.379384</td>\n",
       "      <td>7.381981</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48679 rows Ã— 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM1467273  GSM1467274  GSM1467275  GSM1467276  GSM1467277  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   14.631928   14.846268   14.949614   15.044821   14.790150   \n",
       "ILMN_1343295   13.015539   13.199489   13.256768   13.030246   12.611440   \n",
       "ILMN_1651199    7.407451    7.404480    7.384722    7.398199    7.388903   \n",
       "ILMN_1651209    7.551069    7.489157    7.481180    7.454187    7.468150   \n",
       "ILMN_1651210    7.389822    7.396480    7.392958    7.389106    7.399044   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_2415911    8.200796    8.057956    8.244277    8.318973    8.585545   \n",
       "ILMN_2415926    9.308517    9.261562    9.539295    8.953072    9.178453   \n",
       "ILMN_2415949    8.258738    8.050959    8.363362    8.037019    8.059074   \n",
       "ILMN_2415979   11.414900    9.318214    9.597813    9.561681    9.837778   \n",
       "ILMN_2416019    7.386126    7.390666    7.386612    7.385291    7.383622   \n",
       "\n",
       "              GSM1467278  GSM1467279  GSM1467280  GSM1467281  GSM1467282  ...  \\\n",
       "ID                                                                        ...   \n",
       "ILMN_1343291   14.670395   14.730323   14.972534   14.826768   14.790150  ...   \n",
       "ILMN_1343295   13.030246   12.894845   12.668329   12.536053   12.684084  ...   \n",
       "ILMN_1651199    7.392256    7.411295    7.392441    7.402658    7.397018  ...   \n",
       "ILMN_1651209    7.538856    7.516444    7.428403    7.420693    7.435272  ...   \n",
       "ILMN_1651210    7.401760    7.412553    7.418004    7.383820    7.399614  ...   \n",
       "...                  ...         ...         ...         ...         ...  ...   \n",
       "ILMN_2415911    8.051581    8.212767    8.391308    8.099802    8.173629  ...   \n",
       "ILMN_2415926    9.364314    9.391644    9.428445    9.813872    9.575225  ...   \n",
       "ILMN_2415949    8.154139    8.092218    8.098232    8.231403    8.278169  ...   \n",
       "ILMN_2415979    9.649293   10.331787   10.224846   10.343168   10.534634  ...   \n",
       "ILMN_2416019    7.387200    7.403395    7.388422    7.381764    7.389056  ...   \n",
       "\n",
       "              GSM1467396  GSM1467397  GSM1467398  GSM1467399  GSM1467400  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   14.749395   14.593207   14.558687   14.490661   14.540732   \n",
       "ILMN_1343295   12.729559   12.628114   12.729559   12.526797   12.921417   \n",
       "ILMN_1651199    7.386564    7.392892    7.387466    7.398703    7.387851   \n",
       "ILMN_1651209    7.492618    7.517797    7.570776    7.491987    7.543178   \n",
       "ILMN_1651210    7.383362    7.386934    7.383795    7.390598    7.383517   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_2415911    8.142346    8.187782    8.310146    8.234435    8.165085   \n",
       "ILMN_2415926    9.629529    9.005815    9.073252    9.540430    9.374462   \n",
       "ILMN_2415949    8.041222    7.976685    7.960897    7.998456    8.063397   \n",
       "ILMN_2415979   10.500506   10.723755   10.533912   10.080191   11.034512   \n",
       "ILMN_2416019    7.386483    7.417538    7.380281    7.383249    7.375744   \n",
       "\n",
       "              GSM1467401  GSM1467402  GSM1467403  GSM1467404  GSM1467405  \n",
       "ID                                                                        \n",
       "ILMN_1343291   14.540732   14.972534   14.730323   14.790150   14.609839  \n",
       "ILMN_1343295   12.536053   12.616903   12.685365   12.729559   12.894845  \n",
       "ILMN_1651199    7.380980    7.391920    7.391737    7.384270    7.391287  \n",
       "ILMN_1651209    7.536219    7.486087    7.445162    7.523485    7.524602  \n",
       "ILMN_1651210    7.380075    7.383755    7.392136    7.383855    7.395560  \n",
       "...                  ...         ...         ...         ...         ...  \n",
       "ILMN_2415911    8.222113    7.763713    8.091295    8.353276    7.976434  \n",
       "ILMN_2415926    9.614333    8.026149    8.558921    8.787474    9.170768  \n",
       "ILMN_2415949    7.888010    7.992915    8.265580    7.960649    8.072490  \n",
       "ILMN_2415979   10.533200   11.493401   10.312506   10.355448   10.214869  \n",
       "ILMN_2416019    7.383861    7.384510    7.393648    7.379384    7.381981  \n",
       "\n",
       "[48679 rows x 133 columns]"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_1725881', 'ILMN_1910180', 'ILMN_1804174', 'ILMN_1796063', 'ILMN_1811966'], 'nuID': ['rp13_p1x6D80lNLk3c', 'NEX0oqCV8.er4HVfU4', 'KyqQynMZxJcruyylEU', 'xXl7eXuF7sbPEp.KFI', '9ckqJrioiaej9_ajeQ'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['RefSeq', 'Unigene', 'RefSeq', 'RefSeq', 'RefSeq'], 'Search_Key': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'Transcript': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'ILMN_Gene': ['LOC23117', 'HS.575038', 'FCGR2B', 'TRIM44', 'LOC653895'], 'Source_Reference_ID': ['XM_933824.1', 'Hs.575038', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'RefSeq_ID': ['XM_933824.1', nan, 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Unigene_ID': [nan, 'Hs.575038', nan, nan, nan], 'Entrez_Gene_ID': [23117.0, nan, 2213.0, 54765.0, 653895.0], 'GI': [89040007.0, 10437021.0, 88952550.0, 29029528.0, 89033487.0], 'Accession': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Symbol': ['LOC23117', nan, 'FCGR2B', 'TRIM44', 'LOC653895'], 'Protein_Product': ['XP_938917.1', nan, 'XP_943944.1', 'NP_060053.2', 'XP_941472.1'], 'Array_Address_Id': [1710221.0, 5900364.0, 2480717.0, 1300239.0, 4480719.0], 'Probe_Type': ['I', 'S', 'I', 'S', 'S'], 'Probe_Start': [122.0, 1409.0, 1643.0, 2901.0, 25.0], 'SEQUENCE': ['GGCTCCTCTTTGGGCTCCTACTGGAATTTATCAGCCATCAGTGCATCTCT', 'ACACCTTCAGGAGGGAAGCCCTTATTTCTGGGTTGAACTCCCCTTCCATG', 'TAGGGGCAATAGGCTATACGCTACAGCCTAGGTGTGTAGTAGGCCACACC', 'CCTGCCTGTCTGCCTGTGACCTGTGTACGTATTACAGGCTTTAGGACCAG', 'CTAGCAGGGAGCGGTGAGGGAGAGCGGCTGGATTTCTTGCGGGATCTGCA'], 'Chromosome': ['16', nan, nan, '11', nan], 'Probe_Chr_Orientation': ['-', nan, nan, '+', nan], 'Probe_Coordinates': ['21766363-21766363:21769901-21769949', nan, nan, '35786070-35786119', nan], 'Cytoband': ['16p12.2a', nan, '1q23.3b', '11p13a', '10q11.23b'], 'Definition': ['PREDICTED: Homo sapiens KIAA0220-like protein, transcript variant 11 (LOC23117), mRNA.', 'Homo sapiens cDNA: FLJ21027 fis, clone CAE07110', 'PREDICTED: Homo sapiens Fc fragment of IgG, low affinity IIb, receptor (CD32) (FCGR2B), mRNA.', 'Homo sapiens tripartite motif-containing 44 (TRIM44), mRNA.', 'PREDICTED: Homo sapiens similar to protein geranylgeranyltransferase type I, beta subunit (LOC653895), mRNA.'], 'Ontology_Component': [nan, nan, nan, 'intracellular [goid 5622] [evidence IEA]', nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, 'zinc ion binding [goid 8270] [evidence IEA]; metal ion binding [goid 46872] [evidence IEA]', nan], 'Synonyms': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'Obsolete_Probe_Id': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'GB_ACC': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651199', 'ILMN_1651209', 'ILMN_1651210', 'ILMN_1651221', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651230', 'ILMN_1651232', 'ILMN_1651235', 'ILMN_1651236', 'ILMN_1651237', 'ILMN_1651238', 'ILMN_1651249', 'ILMN_1651253', 'ILMN_1651254', 'ILMN_1651259', 'ILMN_1651260', 'ILMN_1651262']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_1725881', 'ILMN_1910180', 'ILMN_1804174', 'ILMN_1796063', 'ILMN_1811966'], 'nuID': ['rp13_p1x6D80lNLk3c', 'NEX0oqCV8.er4HVfU4', 'KyqQynMZxJcruyylEU', 'xXl7eXuF7sbPEp.KFI', '9ckqJrioiaej9_ajeQ'], 'Species': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Source': ['RefSeq', 'Unigene', 'RefSeq', 'RefSeq', 'RefSeq'], 'Search_Key': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'Transcript': ['ILMN_44919', 'ILMN_127219', 'ILMN_139282', 'ILMN_5006', 'ILMN_38756'], 'ILMN_Gene': ['LOC23117', 'HS.575038', 'FCGR2B', 'TRIM44', 'LOC653895'], 'Source_Reference_ID': ['XM_933824.1', 'Hs.575038', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'RefSeq_ID': ['XM_933824.1', nan, 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Unigene_ID': [nan, 'Hs.575038', nan, nan, nan], 'Entrez_Gene_ID': [23117.0, nan, 2213.0, 54765.0, 653895.0], 'GI': [89040007.0, 10437021.0, 88952550.0, 29029528.0, 89033487.0], 'Accession': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1'], 'Symbol': ['LOC23117', nan, 'FCGR2B', 'TRIM44', 'LOC653895'], 'Protein_Product': ['XP_938917.1', nan, 'XP_943944.1', 'NP_060053.2', 'XP_941472.1'], 'Array_Address_Id': [1710221.0, 5900364.0, 2480717.0, 1300239.0, 4480719.0], 'Probe_Type': ['I', 'S', 'I', 'S', 'S'], 'Probe_Start': [122.0, 1409.0, 1643.0, 2901.0, 25.0], 'SEQUENCE': ['GGCTCCTCTTTGGGCTCCTACTGGAATTTATCAGCCATCAGTGCATCTCT', 'ACACCTTCAGGAGGGAAGCCCTTATTTCTGGGTTGAACTCCCCTTCCATG', 'TAGGGGCAATAGGCTATACGCTACAGCCTAGGTGTGTAGTAGGCCACACC', 'CCTGCCTGTCTGCCTGTGACCTGTGTACGTATTACAGGCTTTAGGACCAG', 'CTAGCAGGGAGCGGTGAGGGAGAGCGGCTGGATTTCTTGCGGGATCTGCA'], 'Chromosome': ['16', nan, nan, '11', nan], 'Probe_Chr_Orientation': ['-', nan, nan, '+', nan], 'Probe_Coordinates': ['21766363-21766363:21769901-21769949', nan, nan, '35786070-35786119', nan], 'Cytoband': ['16p12.2a', nan, '1q23.3b', '11p13a', '10q11.23b'], 'Definition': ['PREDICTED: Homo sapiens KIAA0220-like protein, transcript variant 11 (LOC23117), mRNA.', 'Homo sapiens cDNA: FLJ21027 fis, clone CAE07110', 'PREDICTED: Homo sapiens Fc fragment of IgG, low affinity IIb, receptor (CD32) (FCGR2B), mRNA.', 'Homo sapiens tripartite motif-containing 44 (TRIM44), mRNA.', 'PREDICTED: Homo sapiens similar to protein geranylgeranyltransferase type I, beta subunit (LOC653895), mRNA.'], 'Ontology_Component': [nan, nan, nan, 'intracellular [goid 5622] [evidence IEA]', nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, 'zinc ion binding [goid 8270] [evidence IEA]; metal ion binding [goid 46872] [evidence IEA]', nan], 'Synonyms': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'Obsolete_Probe_Id': [nan, nan, nan, 'MGC3490; MC7; HSA249128; DIPB', nan], 'GB_ACC': ['XM_933824.1', 'AK024680', 'XM_938851.1', 'NM_017583.3', 'XM_936379.1']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'nuID', 'Species', 'Source', 'Search_Key', 'Transcript',\n",
       "       'ILMN_Gene', 'Source_Reference_ID', 'RefSeq_ID', 'Unigene_ID',\n",
       "       'Entrez_Gene_ID', 'GI', 'Accession', 'Symbol', 'Protein_Product',\n",
       "       'Array_Address_Id', 'Probe_Type', 'Probe_Start', 'SEQUENCE',\n",
       "       'Chromosome', 'Probe_Chr_Orientation', 'Probe_Coordinates', 'Cytoband',\n",
       "       'Definition', 'Ontology_Component', 'Ontology_Process',\n",
       "       'Ontology_Function', 'Synonyms', 'Obsolete_Probe_Id', 'GB_ACC'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1467273</th>\n",
       "      <th>GSM1467274</th>\n",
       "      <th>GSM1467275</th>\n",
       "      <th>GSM1467276</th>\n",
       "      <th>GSM1467277</th>\n",
       "      <th>GSM1467278</th>\n",
       "      <th>GSM1467279</th>\n",
       "      <th>GSM1467280</th>\n",
       "      <th>GSM1467281</th>\n",
       "      <th>GSM1467282</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1467396</th>\n",
       "      <th>GSM1467397</th>\n",
       "      <th>GSM1467398</th>\n",
       "      <th>GSM1467399</th>\n",
       "      <th>GSM1467400</th>\n",
       "      <th>GSM1467401</th>\n",
       "      <th>GSM1467402</th>\n",
       "      <th>GSM1467403</th>\n",
       "      <th>GSM1467404</th>\n",
       "      <th>GSM1467405</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>7.481047</td>\n",
       "      <td>7.459438</td>\n",
       "      <td>7.462799</td>\n",
       "      <td>7.453585</td>\n",
       "      <td>7.434372</td>\n",
       "      <td>7.435308</td>\n",
       "      <td>7.497011</td>\n",
       "      <td>7.434371</td>\n",
       "      <td>7.413374</td>\n",
       "      <td>7.439196</td>\n",
       "      <td>...</td>\n",
       "      <td>7.497660</td>\n",
       "      <td>7.457713</td>\n",
       "      <td>7.427958</td>\n",
       "      <td>7.443941</td>\n",
       "      <td>7.438771</td>\n",
       "      <td>7.455120</td>\n",
       "      <td>7.419787</td>\n",
       "      <td>7.442493</td>\n",
       "      <td>7.427613</td>\n",
       "      <td>7.424304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>7.405821</td>\n",
       "      <td>7.396154</td>\n",
       "      <td>7.399692</td>\n",
       "      <td>7.411053</td>\n",
       "      <td>7.427684</td>\n",
       "      <td>7.420371</td>\n",
       "      <td>7.409882</td>\n",
       "      <td>7.406598</td>\n",
       "      <td>7.423508</td>\n",
       "      <td>7.403380</td>\n",
       "      <td>...</td>\n",
       "      <td>7.420024</td>\n",
       "      <td>7.425435</td>\n",
       "      <td>7.405370</td>\n",
       "      <td>7.404043</td>\n",
       "      <td>7.407866</td>\n",
       "      <td>7.410821</td>\n",
       "      <td>7.398595</td>\n",
       "      <td>7.412058</td>\n",
       "      <td>7.404174</td>\n",
       "      <td>7.397455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>8.369617</td>\n",
       "      <td>8.115136</td>\n",
       "      <td>7.834904</td>\n",
       "      <td>8.268345</td>\n",
       "      <td>8.132207</td>\n",
       "      <td>8.514582</td>\n",
       "      <td>8.681611</td>\n",
       "      <td>8.820405</td>\n",
       "      <td>8.701411</td>\n",
       "      <td>8.446774</td>\n",
       "      <td>...</td>\n",
       "      <td>8.942798</td>\n",
       "      <td>8.940076</td>\n",
       "      <td>8.665946</td>\n",
       "      <td>8.740932</td>\n",
       "      <td>8.765340</td>\n",
       "      <td>8.901560</td>\n",
       "      <td>9.056649</td>\n",
       "      <td>9.500376</td>\n",
       "      <td>8.691591</td>\n",
       "      <td>8.890048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>7.381102</td>\n",
       "      <td>7.424990</td>\n",
       "      <td>7.406004</td>\n",
       "      <td>7.397059</td>\n",
       "      <td>7.408826</td>\n",
       "      <td>7.394649</td>\n",
       "      <td>7.410082</td>\n",
       "      <td>7.443830</td>\n",
       "      <td>7.402433</td>\n",
       "      <td>7.527460</td>\n",
       "      <td>...</td>\n",
       "      <td>7.387483</td>\n",
       "      <td>7.386391</td>\n",
       "      <td>7.434159</td>\n",
       "      <td>7.406756</td>\n",
       "      <td>7.388920</td>\n",
       "      <td>7.403031</td>\n",
       "      <td>7.630691</td>\n",
       "      <td>7.391827</td>\n",
       "      <td>7.423193</td>\n",
       "      <td>7.387081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>7.395020</td>\n",
       "      <td>7.384147</td>\n",
       "      <td>7.385292</td>\n",
       "      <td>7.387125</td>\n",
       "      <td>7.389030</td>\n",
       "      <td>7.395563</td>\n",
       "      <td>7.393727</td>\n",
       "      <td>7.381826</td>\n",
       "      <td>7.391831</td>\n",
       "      <td>7.383787</td>\n",
       "      <td>...</td>\n",
       "      <td>7.381678</td>\n",
       "      <td>7.383456</td>\n",
       "      <td>7.390774</td>\n",
       "      <td>7.385387</td>\n",
       "      <td>7.381505</td>\n",
       "      <td>7.391571</td>\n",
       "      <td>7.393197</td>\n",
       "      <td>7.380317</td>\n",
       "      <td>7.390618</td>\n",
       "      <td>7.384771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>11.755243</td>\n",
       "      <td>11.580051</td>\n",
       "      <td>11.472778</td>\n",
       "      <td>11.197515</td>\n",
       "      <td>11.357074</td>\n",
       "      <td>11.292039</td>\n",
       "      <td>11.431473</td>\n",
       "      <td>11.529512</td>\n",
       "      <td>11.584090</td>\n",
       "      <td>11.731810</td>\n",
       "      <td>...</td>\n",
       "      <td>11.920459</td>\n",
       "      <td>11.946019</td>\n",
       "      <td>11.600578</td>\n",
       "      <td>11.121539</td>\n",
       "      <td>11.563971</td>\n",
       "      <td>11.461402</td>\n",
       "      <td>11.607571</td>\n",
       "      <td>11.942620</td>\n",
       "      <td>12.143056</td>\n",
       "      <td>11.889138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>9.488777</td>\n",
       "      <td>8.958607</td>\n",
       "      <td>8.853385</td>\n",
       "      <td>9.023618</td>\n",
       "      <td>9.076320</td>\n",
       "      <td>8.989631</td>\n",
       "      <td>8.968002</td>\n",
       "      <td>8.859351</td>\n",
       "      <td>9.026985</td>\n",
       "      <td>9.067510</td>\n",
       "      <td>...</td>\n",
       "      <td>8.818570</td>\n",
       "      <td>9.242392</td>\n",
       "      <td>9.202663</td>\n",
       "      <td>9.317282</td>\n",
       "      <td>9.217690</td>\n",
       "      <td>9.172183</td>\n",
       "      <td>9.902353</td>\n",
       "      <td>8.986207</td>\n",
       "      <td>9.157025</td>\n",
       "      <td>9.066989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>9.066159</td>\n",
       "      <td>8.598233</td>\n",
       "      <td>8.774467</td>\n",
       "      <td>8.956653</td>\n",
       "      <td>8.843062</td>\n",
       "      <td>9.069212</td>\n",
       "      <td>8.888722</td>\n",
       "      <td>8.928407</td>\n",
       "      <td>9.072954</td>\n",
       "      <td>8.852790</td>\n",
       "      <td>...</td>\n",
       "      <td>8.847695</td>\n",
       "      <td>8.790686</td>\n",
       "      <td>8.876767</td>\n",
       "      <td>9.085703</td>\n",
       "      <td>8.807783</td>\n",
       "      <td>8.747593</td>\n",
       "      <td>8.930893</td>\n",
       "      <td>8.798290</td>\n",
       "      <td>8.892741</td>\n",
       "      <td>8.587646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>8.523295</td>\n",
       "      <td>8.584810</td>\n",
       "      <td>8.731412</td>\n",
       "      <td>8.739782</td>\n",
       "      <td>8.744687</td>\n",
       "      <td>8.432901</td>\n",
       "      <td>8.545003</td>\n",
       "      <td>8.718865</td>\n",
       "      <td>8.629061</td>\n",
       "      <td>8.707885</td>\n",
       "      <td>...</td>\n",
       "      <td>8.760058</td>\n",
       "      <td>8.684442</td>\n",
       "      <td>8.750618</td>\n",
       "      <td>8.786311</td>\n",
       "      <td>8.544695</td>\n",
       "      <td>8.787504</td>\n",
       "      <td>8.753251</td>\n",
       "      <td>8.602089</td>\n",
       "      <td>8.709299</td>\n",
       "      <td>8.472725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB1C</th>\n",
       "      <td>7.408587</td>\n",
       "      <td>7.403482</td>\n",
       "      <td>7.377490</td>\n",
       "      <td>7.383207</td>\n",
       "      <td>7.390069</td>\n",
       "      <td>7.397665</td>\n",
       "      <td>7.387363</td>\n",
       "      <td>7.380755</td>\n",
       "      <td>7.387831</td>\n",
       "      <td>7.383861</td>\n",
       "      <td>...</td>\n",
       "      <td>7.390564</td>\n",
       "      <td>7.385140</td>\n",
       "      <td>7.403709</td>\n",
       "      <td>7.401325</td>\n",
       "      <td>7.387384</td>\n",
       "      <td>7.389036</td>\n",
       "      <td>7.382404</td>\n",
       "      <td>7.406016</td>\n",
       "      <td>7.383361</td>\n",
       "      <td>7.412470</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15589 rows Ã— 133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1467273  GSM1467274  GSM1467275  GSM1467276  GSM1467277  \\\n",
       "A1BG       7.481047    7.459438    7.462799    7.453585    7.434372   \n",
       "A1CF       7.405821    7.396154    7.399692    7.411053    7.427684   \n",
       "A2M        8.369617    8.115136    7.834904    8.268345    8.132207   \n",
       "A2ML1      7.381102    7.424990    7.406004    7.397059    7.408826   \n",
       "A3GALT2    7.395020    7.384147    7.385292    7.387125    7.389030   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B    11.755243   11.580051   11.472778   11.197515   11.357074   \n",
       "ZYX        9.488777    8.958607    8.853385    9.023618    9.076320   \n",
       "ZZEF1      9.066159    8.598233    8.774467    8.956653    8.843062   \n",
       "ZZZ3       8.523295    8.584810    8.731412    8.739782    8.744687   \n",
       "RAB1C      7.408587    7.403482    7.377490    7.383207    7.390069   \n",
       "\n",
       "         GSM1467278  GSM1467279  GSM1467280  GSM1467281  GSM1467282  ...  \\\n",
       "A1BG       7.435308    7.497011    7.434371    7.413374    7.439196  ...   \n",
       "A1CF       7.420371    7.409882    7.406598    7.423508    7.403380  ...   \n",
       "A2M        8.514582    8.681611    8.820405    8.701411    8.446774  ...   \n",
       "A2ML1      7.394649    7.410082    7.443830    7.402433    7.527460  ...   \n",
       "A3GALT2    7.395563    7.393727    7.381826    7.391831    7.383787  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZYG11B    11.292039   11.431473   11.529512   11.584090   11.731810  ...   \n",
       "ZYX        8.989631    8.968002    8.859351    9.026985    9.067510  ...   \n",
       "ZZEF1      9.069212    8.888722    8.928407    9.072954    8.852790  ...   \n",
       "ZZZ3       8.432901    8.545003    8.718865    8.629061    8.707885  ...   \n",
       "RAB1C      7.397665    7.387363    7.380755    7.387831    7.383861  ...   \n",
       "\n",
       "         GSM1467396  GSM1467397  GSM1467398  GSM1467399  GSM1467400  \\\n",
       "A1BG       7.497660    7.457713    7.427958    7.443941    7.438771   \n",
       "A1CF       7.420024    7.425435    7.405370    7.404043    7.407866   \n",
       "A2M        8.942798    8.940076    8.665946    8.740932    8.765340   \n",
       "A2ML1      7.387483    7.386391    7.434159    7.406756    7.388920   \n",
       "A3GALT2    7.381678    7.383456    7.390774    7.385387    7.381505   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B    11.920459   11.946019   11.600578   11.121539   11.563971   \n",
       "ZYX        8.818570    9.242392    9.202663    9.317282    9.217690   \n",
       "ZZEF1      8.847695    8.790686    8.876767    9.085703    8.807783   \n",
       "ZZZ3       8.760058    8.684442    8.750618    8.786311    8.544695   \n",
       "RAB1C      7.390564    7.385140    7.403709    7.401325    7.387384   \n",
       "\n",
       "         GSM1467401  GSM1467402  GSM1467403  GSM1467404  GSM1467405  \n",
       "A1BG       7.455120    7.419787    7.442493    7.427613    7.424304  \n",
       "A1CF       7.410821    7.398595    7.412058    7.404174    7.397455  \n",
       "A2M        8.901560    9.056649    9.500376    8.691591    8.890048  \n",
       "A2ML1      7.403031    7.630691    7.391827    7.423193    7.387081  \n",
       "A3GALT2    7.391571    7.393197    7.380317    7.390618    7.384771  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZYG11B    11.461402   11.607571   11.942620   12.143056   11.889138  \n",
       "ZYX        9.172183    9.902353    8.986207    9.157025    9.066989  \n",
       "ZZEF1      8.747593    8.930893    8.798290    8.892741    8.587646  \n",
       "ZZZ3       8.787504    8.753251    8.602089    8.709299    8.472725  \n",
       "RAB1C      7.389036    7.382404    7.406016    7.383361    7.412470  \n",
       "\n",
       "[15589 rows x 133 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>RAB1C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1467273</th>\n",
       "      <td>0.0</td>\n",
       "      <td>50.42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.481047</td>\n",
       "      <td>7.405821</td>\n",
       "      <td>8.369617</td>\n",
       "      <td>7.381102</td>\n",
       "      <td>7.395020</td>\n",
       "      <td>7.445335</td>\n",
       "      <td>7.418757</td>\n",
       "      <td>...</td>\n",
       "      <td>7.422142</td>\n",
       "      <td>7.411947</td>\n",
       "      <td>7.688330</td>\n",
       "      <td>7.692514</td>\n",
       "      <td>7.397040</td>\n",
       "      <td>11.755243</td>\n",
       "      <td>9.488777</td>\n",
       "      <td>9.066159</td>\n",
       "      <td>8.523295</td>\n",
       "      <td>7.408587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467274</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.459438</td>\n",
       "      <td>7.396154</td>\n",
       "      <td>8.115136</td>\n",
       "      <td>7.424990</td>\n",
       "      <td>7.384147</td>\n",
       "      <td>7.413263</td>\n",
       "      <td>7.399258</td>\n",
       "      <td>...</td>\n",
       "      <td>7.438231</td>\n",
       "      <td>7.418272</td>\n",
       "      <td>7.734598</td>\n",
       "      <td>7.639503</td>\n",
       "      <td>7.393080</td>\n",
       "      <td>11.580051</td>\n",
       "      <td>8.958607</td>\n",
       "      <td>8.598233</td>\n",
       "      <td>8.584810</td>\n",
       "      <td>7.403482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467275</th>\n",
       "      <td>1.0</td>\n",
       "      <td>30.63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.462799</td>\n",
       "      <td>7.399692</td>\n",
       "      <td>7.834904</td>\n",
       "      <td>7.406004</td>\n",
       "      <td>7.385292</td>\n",
       "      <td>7.481972</td>\n",
       "      <td>7.389239</td>\n",
       "      <td>...</td>\n",
       "      <td>7.442823</td>\n",
       "      <td>7.420557</td>\n",
       "      <td>7.897464</td>\n",
       "      <td>7.799014</td>\n",
       "      <td>7.396474</td>\n",
       "      <td>11.472778</td>\n",
       "      <td>8.853385</td>\n",
       "      <td>8.774467</td>\n",
       "      <td>8.731412</td>\n",
       "      <td>7.377490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467276</th>\n",
       "      <td>0.0</td>\n",
       "      <td>61.17</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.453585</td>\n",
       "      <td>7.411053</td>\n",
       "      <td>8.268345</td>\n",
       "      <td>7.397059</td>\n",
       "      <td>7.387125</td>\n",
       "      <td>7.463897</td>\n",
       "      <td>7.401978</td>\n",
       "      <td>...</td>\n",
       "      <td>7.415354</td>\n",
       "      <td>7.414003</td>\n",
       "      <td>7.809488</td>\n",
       "      <td>7.719431</td>\n",
       "      <td>7.408978</td>\n",
       "      <td>11.197515</td>\n",
       "      <td>9.023618</td>\n",
       "      <td>8.956653</td>\n",
       "      <td>8.739782</td>\n",
       "      <td>7.383207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467277</th>\n",
       "      <td>0.0</td>\n",
       "      <td>32.70</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.434372</td>\n",
       "      <td>7.427684</td>\n",
       "      <td>8.132207</td>\n",
       "      <td>7.408826</td>\n",
       "      <td>7.389030</td>\n",
       "      <td>7.414156</td>\n",
       "      <td>7.415771</td>\n",
       "      <td>...</td>\n",
       "      <td>7.389650</td>\n",
       "      <td>7.418803</td>\n",
       "      <td>7.653474</td>\n",
       "      <td>7.805373</td>\n",
       "      <td>7.393036</td>\n",
       "      <td>11.357074</td>\n",
       "      <td>9.076320</td>\n",
       "      <td>8.843062</td>\n",
       "      <td>8.744687</td>\n",
       "      <td>7.390069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467401</th>\n",
       "      <td>0.0</td>\n",
       "      <td>44.96</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.455120</td>\n",
       "      <td>7.410821</td>\n",
       "      <td>8.901560</td>\n",
       "      <td>7.403031</td>\n",
       "      <td>7.391571</td>\n",
       "      <td>7.433895</td>\n",
       "      <td>7.459833</td>\n",
       "      <td>...</td>\n",
       "      <td>7.427711</td>\n",
       "      <td>7.424786</td>\n",
       "      <td>7.672185</td>\n",
       "      <td>7.568192</td>\n",
       "      <td>7.391293</td>\n",
       "      <td>11.461402</td>\n",
       "      <td>9.172183</td>\n",
       "      <td>8.747593</td>\n",
       "      <td>8.787504</td>\n",
       "      <td>7.389036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467402</th>\n",
       "      <td>0.0</td>\n",
       "      <td>35.68</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.419787</td>\n",
       "      <td>7.398595</td>\n",
       "      <td>9.056649</td>\n",
       "      <td>7.630691</td>\n",
       "      <td>7.393197</td>\n",
       "      <td>8.135820</td>\n",
       "      <td>7.411807</td>\n",
       "      <td>...</td>\n",
       "      <td>7.399261</td>\n",
       "      <td>7.415387</td>\n",
       "      <td>7.559787</td>\n",
       "      <td>7.702635</td>\n",
       "      <td>7.389805</td>\n",
       "      <td>11.607571</td>\n",
       "      <td>9.902353</td>\n",
       "      <td>8.930893</td>\n",
       "      <td>8.753251</td>\n",
       "      <td>7.382404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467403</th>\n",
       "      <td>0.0</td>\n",
       "      <td>63.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.442493</td>\n",
       "      <td>7.412058</td>\n",
       "      <td>9.500376</td>\n",
       "      <td>7.391827</td>\n",
       "      <td>7.380317</td>\n",
       "      <td>7.532850</td>\n",
       "      <td>7.447209</td>\n",
       "      <td>...</td>\n",
       "      <td>7.415293</td>\n",
       "      <td>7.420746</td>\n",
       "      <td>7.617346</td>\n",
       "      <td>7.715541</td>\n",
       "      <td>7.388253</td>\n",
       "      <td>11.942620</td>\n",
       "      <td>8.986207</td>\n",
       "      <td>8.798290</td>\n",
       "      <td>8.602089</td>\n",
       "      <td>7.406016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467404</th>\n",
       "      <td>0.0</td>\n",
       "      <td>38.78</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.427613</td>\n",
       "      <td>7.404174</td>\n",
       "      <td>8.691591</td>\n",
       "      <td>7.423193</td>\n",
       "      <td>7.390618</td>\n",
       "      <td>7.440411</td>\n",
       "      <td>7.416510</td>\n",
       "      <td>...</td>\n",
       "      <td>7.412564</td>\n",
       "      <td>7.397225</td>\n",
       "      <td>7.710904</td>\n",
       "      <td>7.607852</td>\n",
       "      <td>7.389100</td>\n",
       "      <td>12.143056</td>\n",
       "      <td>9.157025</td>\n",
       "      <td>8.892741</td>\n",
       "      <td>8.709299</td>\n",
       "      <td>7.383361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1467405</th>\n",
       "      <td>0.0</td>\n",
       "      <td>45.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.424304</td>\n",
       "      <td>7.397455</td>\n",
       "      <td>8.890048</td>\n",
       "      <td>7.387081</td>\n",
       "      <td>7.384771</td>\n",
       "      <td>7.419501</td>\n",
       "      <td>7.415334</td>\n",
       "      <td>...</td>\n",
       "      <td>7.417651</td>\n",
       "      <td>7.399305</td>\n",
       "      <td>7.719476</td>\n",
       "      <td>7.560402</td>\n",
       "      <td>7.388926</td>\n",
       "      <td>11.889138</td>\n",
       "      <td>9.066989</td>\n",
       "      <td>8.587646</td>\n",
       "      <td>8.472725</td>\n",
       "      <td>7.412470</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133 rows Ã— 15592 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Anxiety disorder    Age  Gender      A1BG      A1CF       A2M  \\\n",
       "GSM1467273               0.0  50.42     1.0  7.481047  7.405821  8.369617   \n",
       "GSM1467274               0.0  27.50     0.0  7.459438  7.396154  8.115136   \n",
       "GSM1467275               1.0  30.63     0.0  7.462799  7.399692  7.834904   \n",
       "GSM1467276               0.0  61.17     0.0  7.453585  7.411053  8.268345   \n",
       "GSM1467277               0.0  32.70     0.0  7.434372  7.427684  8.132207   \n",
       "...                      ...    ...     ...       ...       ...       ...   \n",
       "GSM1467401               0.0  44.96     0.0  7.455120  7.410821  8.901560   \n",
       "GSM1467402               0.0  35.68     1.0  7.419787  7.398595  9.056649   \n",
       "GSM1467403               0.0  63.00     0.0  7.442493  7.412058  9.500376   \n",
       "GSM1467404               0.0  38.78     0.0  7.427613  7.404174  8.691591   \n",
       "GSM1467405               0.0  45.98     0.0  7.424304  7.397455  8.890048   \n",
       "\n",
       "               A2ML1   A3GALT2    A4GALT     A4GNT  ...     ZWINT      ZXDA  \\\n",
       "GSM1467273  7.381102  7.395020  7.445335  7.418757  ...  7.422142  7.411947   \n",
       "GSM1467274  7.424990  7.384147  7.413263  7.399258  ...  7.438231  7.418272   \n",
       "GSM1467275  7.406004  7.385292  7.481972  7.389239  ...  7.442823  7.420557   \n",
       "GSM1467276  7.397059  7.387125  7.463897  7.401978  ...  7.415354  7.414003   \n",
       "GSM1467277  7.408826  7.389030  7.414156  7.415771  ...  7.389650  7.418803   \n",
       "...              ...       ...       ...       ...  ...       ...       ...   \n",
       "GSM1467401  7.403031  7.391571  7.433895  7.459833  ...  7.427711  7.424786   \n",
       "GSM1467402  7.630691  7.393197  8.135820  7.411807  ...  7.399261  7.415387   \n",
       "GSM1467403  7.391827  7.380317  7.532850  7.447209  ...  7.415293  7.420746   \n",
       "GSM1467404  7.423193  7.390618  7.440411  7.416510  ...  7.412564  7.397225   \n",
       "GSM1467405  7.387081  7.384771  7.419501  7.415334  ...  7.417651  7.399305   \n",
       "\n",
       "                ZXDB      ZXDC    ZYG11A     ZYG11B       ZYX     ZZEF1  \\\n",
       "GSM1467273  7.688330  7.692514  7.397040  11.755243  9.488777  9.066159   \n",
       "GSM1467274  7.734598  7.639503  7.393080  11.580051  8.958607  8.598233   \n",
       "GSM1467275  7.897464  7.799014  7.396474  11.472778  8.853385  8.774467   \n",
       "GSM1467276  7.809488  7.719431  7.408978  11.197515  9.023618  8.956653   \n",
       "GSM1467277  7.653474  7.805373  7.393036  11.357074  9.076320  8.843062   \n",
       "...              ...       ...       ...        ...       ...       ...   \n",
       "GSM1467401  7.672185  7.568192  7.391293  11.461402  9.172183  8.747593   \n",
       "GSM1467402  7.559787  7.702635  7.389805  11.607571  9.902353  8.930893   \n",
       "GSM1467403  7.617346  7.715541  7.388253  11.942620  8.986207  8.798290   \n",
       "GSM1467404  7.710904  7.607852  7.389100  12.143056  9.157025  8.892741   \n",
       "GSM1467405  7.719476  7.560402  7.388926  11.889138  9.066989  8.587646   \n",
       "\n",
       "                ZZZ3     RAB1C  \n",
       "GSM1467273  8.523295  7.408587  \n",
       "GSM1467274  8.584810  7.403482  \n",
       "GSM1467275  8.731412  7.377490  \n",
       "GSM1467276  8.739782  7.383207  \n",
       "GSM1467277  8.744687  7.390069  \n",
       "...              ...       ...  \n",
       "GSM1467401  8.787504  7.389036  \n",
       "GSM1467402  8.753251  7.382404  \n",
       "GSM1467403  8.602089  7.406016  \n",
       "GSM1467404  8.709299  7.383361  \n",
       "GSM1467405  8.472725  7.412470  \n",
       "\n",
       "[133 rows x 15592 columns]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 133 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1.0' with 16 occurrences. This represents 12.03% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 30.04\n",
      "  50% (Median): 45.22\n",
      "  75%: 54.42\n",
      "Min: 16.18\n",
      "Max: 84.06\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 36 occurrences. This represents 27.07% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"lncRNA and mRNA expression data in peripheral blood sampled from patients with Obsessive-Compulsive Disorder\"\n",
      "!Series_summary\t\"The aim of the study is to identify the global messenger RNA (mRNA) and long noncoding RNA (lncRNA) expression profiling in peripheral blood from thirty patients with Obsessive Compulsive Disorders (OCD) and thirty paired normal controls.\"\n",
      "!Series_overall_design\t\"We quantified the gene transcripts in peripheral blood from thirty patients with OCD and thirty normal controls by the method of Microarray using Aglilent G3 lncRNA v4.04Ã—180K.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2067403</th>\n",
       "      <th>GSM2067404</th>\n",
       "      <th>GSM2067405</th>\n",
       "      <th>GSM2067406</th>\n",
       "      <th>GSM2067407</th>\n",
       "      <th>GSM2067408</th>\n",
       "      <th>GSM2067409</th>\n",
       "      <th>GSM2067410</th>\n",
       "      <th>GSM2067411</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2067453</th>\n",
       "      <th>GSM2067454</th>\n",
       "      <th>GSM2067455</th>\n",
       "      <th>GSM2067456</th>\n",
       "      <th>GSM2067457</th>\n",
       "      <th>GSM2067458</th>\n",
       "      <th>GSM2067459</th>\n",
       "      <th>GSM2067460</th>\n",
       "      <th>GSM2067461</th>\n",
       "      <th>GSM2067462</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>disease state: Obsessive-Compulsive Disorder</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "      <td>disease state: normal control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 25y</td>\n",
       "      <td>age: 23y</td>\n",
       "      <td>age: 18y</td>\n",
       "      <td>age: 26y</td>\n",
       "      <td>age: 27y</td>\n",
       "      <td>age: 19y</td>\n",
       "      <td>age: 22y</td>\n",
       "      <td>age: 27y</td>\n",
       "      <td>age: 18y</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 43y</td>\n",
       "      <td>age: 40y</td>\n",
       "      <td>age: 32y</td>\n",
       "      <td>age: 28y</td>\n",
       "      <td>age: 27y</td>\n",
       "      <td>age: 30y</td>\n",
       "      <td>age: 24y</td>\n",
       "      <td>age: 35y</td>\n",
       "      <td>age: 56y</td>\n",
       "      <td>age: 56y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows Ã— 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                    GSM2067403  \\\n",
       "0  !Sample_characteristics_ch1                           tissue: whole blood   \n",
       "1  !Sample_characteristics_ch1  disease state: Obsessive-Compulsive Disorder   \n",
       "2  !Sample_characteristics_ch1                                  gender: male   \n",
       "3  !Sample_characteristics_ch1                                      age: 25y   \n",
       "\n",
       "                                     GSM2067404  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                gender: female   \n",
       "3                                      age: 23y   \n",
       "\n",
       "                                     GSM2067405  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                gender: female   \n",
       "3                                      age: 18y   \n",
       "\n",
       "                                     GSM2067406  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                  gender: male   \n",
       "3                                      age: 26y   \n",
       "\n",
       "                                     GSM2067407  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                gender: female   \n",
       "3                                      age: 27y   \n",
       "\n",
       "                                     GSM2067408  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                  gender: male   \n",
       "3                                      age: 19y   \n",
       "\n",
       "                                     GSM2067409  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                  gender: male   \n",
       "3                                      age: 22y   \n",
       "\n",
       "                                     GSM2067410  \\\n",
       "0                           tissue: whole blood   \n",
       "1  disease state: Obsessive-Compulsive Disorder   \n",
       "2                                gender: female   \n",
       "3                                      age: 27y   \n",
       "\n",
       "                                     GSM2067411  ...  \\\n",
       "0                           tissue: whole blood  ...   \n",
       "1  disease state: Obsessive-Compulsive Disorder  ...   \n",
       "2                                  gender: male  ...   \n",
       "3                                      age: 18y  ...   \n",
       "\n",
       "                      GSM2067453                     GSM2067454  \\\n",
       "0            tissue: whole blood            tissue: whole blood   \n",
       "1  disease state: normal control  disease state: normal control   \n",
       "2                 gender: female                   gender: male   \n",
       "3                       age: 43y                       age: 40y   \n",
       "\n",
       "                      GSM2067455                     GSM2067456  \\\n",
       "0            tissue: whole blood            tissue: whole blood   \n",
       "1  disease state: normal control  disease state: normal control   \n",
       "2                   gender: male                 gender: female   \n",
       "3                       age: 32y                       age: 28y   \n",
       "\n",
       "                      GSM2067457                     GSM2067458  \\\n",
       "0            tissue: whole blood            tissue: whole blood   \n",
       "1  disease state: normal control  disease state: normal control   \n",
       "2                   gender: male                   gender: male   \n",
       "3                       age: 27y                       age: 30y   \n",
       "\n",
       "                      GSM2067459                     GSM2067460  \\\n",
       "0            tissue: whole blood            tissue: whole blood   \n",
       "1  disease state: normal control  disease state: normal control   \n",
       "2                   gender: male                 gender: female   \n",
       "3                       age: 24y                       age: 35y   \n",
       "\n",
       "                      GSM2067461                     GSM2067462  \n",
       "0            tissue: whole blood            tissue: whole blood  \n",
       "1  disease state: normal control  disease state: normal control  \n",
       "2                   gender: male                 gender: female  \n",
       "3                       age: 56y                       age: 56y  \n",
       "\n",
       "[4 rows x 61 columns]"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE78104\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'disease state: Obsessive-Compulsive Disorder',\n",
       "       'disease state: normal control'], dtype=object)"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 3\n",
    "gender_row = 2\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease state: Obsessive-Compulsive Disorder':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "    \n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1].rstrip('y'))\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2067403</th>\n",
       "      <th>GSM2067404</th>\n",
       "      <th>GSM2067405</th>\n",
       "      <th>GSM2067406</th>\n",
       "      <th>GSM2067407</th>\n",
       "      <th>GSM2067408</th>\n",
       "      <th>GSM2067409</th>\n",
       "      <th>GSM2067410</th>\n",
       "      <th>GSM2067411</th>\n",
       "      <th>GSM2067412</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2067453</th>\n",
       "      <th>GSM2067454</th>\n",
       "      <th>GSM2067455</th>\n",
       "      <th>GSM2067456</th>\n",
       "      <th>GSM2067457</th>\n",
       "      <th>GSM2067458</th>\n",
       "      <th>GSM2067459</th>\n",
       "      <th>GSM2067460</th>\n",
       "      <th>GSM2067461</th>\n",
       "      <th>GSM2067462</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>25</td>\n",
       "      <td>23</td>\n",
       "      <td>18</td>\n",
       "      <td>26</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>27</td>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>43</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>35</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM2067403  GSM2067404  GSM2067405  GSM2067406  GSM2067407  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Age                       25          23          18          26          27   \n",
       "Gender                     0           1           1           0           1   \n",
       "\n",
       "                  GSM2067408  GSM2067409  GSM2067410  GSM2067411  GSM2067412  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Age                       19          22          27          18          25   \n",
       "Gender                     0           0           1           0           0   \n",
       "\n",
       "                  ...  GSM2067453  GSM2067454  GSM2067455  GSM2067456  \\\n",
       "Anxiety disorder  ...           0           0           0           0   \n",
       "Age               ...          43          40          32          28   \n",
       "Gender            ...           1           0           0           1   \n",
       "\n",
       "                  GSM2067457  GSM2067458  GSM2067459  GSM2067460  GSM2067461  \\\n",
       "Anxiety disorder           0           0           0           0           0   \n",
       "Age                       27          30          24          35          56   \n",
       "Gender                     0           0           0           1           0   \n",
       "\n",
       "                  GSM2067462  \n",
       "Anxiety disorder           0  \n",
       "Age                       56  \n",
       "Gender                     1  \n",
       "\n",
       "[3 rows x 60 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2067403</th>\n",
       "      <th>GSM2067404</th>\n",
       "      <th>GSM2067405</th>\n",
       "      <th>GSM2067406</th>\n",
       "      <th>GSM2067407</th>\n",
       "      <th>GSM2067408</th>\n",
       "      <th>GSM2067409</th>\n",
       "      <th>GSM2067410</th>\n",
       "      <th>GSM2067411</th>\n",
       "      <th>GSM2067412</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2067453</th>\n",
       "      <th>GSM2067454</th>\n",
       "      <th>GSM2067455</th>\n",
       "      <th>GSM2067456</th>\n",
       "      <th>GSM2067457</th>\n",
       "      <th>GSM2067458</th>\n",
       "      <th>GSM2067459</th>\n",
       "      <th>GSM2067460</th>\n",
       "      <th>GSM2067461</th>\n",
       "      <th>GSM2067462</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(+)E1A_r60_1</th>\n",
       "      <td>13.442002</td>\n",
       "      <td>13.263885</td>\n",
       "      <td>13.225509</td>\n",
       "      <td>13.144137</td>\n",
       "      <td>12.986026</td>\n",
       "      <td>13.849309</td>\n",
       "      <td>13.541306</td>\n",
       "      <td>14.152843</td>\n",
       "      <td>13.809382</td>\n",
       "      <td>14.053410</td>\n",
       "      <td>...</td>\n",
       "      <td>13.750706</td>\n",
       "      <td>13.790617</td>\n",
       "      <td>13.543028</td>\n",
       "      <td>13.721816</td>\n",
       "      <td>14.025154</td>\n",
       "      <td>14.458712</td>\n",
       "      <td>14.075050</td>\n",
       "      <td>14.186631</td>\n",
       "      <td>14.150471</td>\n",
       "      <td>13.071623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(+)E1A_r60_3</th>\n",
       "      <td>5.764329</td>\n",
       "      <td>6.253843</td>\n",
       "      <td>6.890254</td>\n",
       "      <td>6.576528</td>\n",
       "      <td>6.647077</td>\n",
       "      <td>7.138355</td>\n",
       "      <td>7.172650</td>\n",
       "      <td>7.322565</td>\n",
       "      <td>6.515238</td>\n",
       "      <td>6.263576</td>\n",
       "      <td>...</td>\n",
       "      <td>6.291416</td>\n",
       "      <td>6.574324</td>\n",
       "      <td>6.940987</td>\n",
       "      <td>6.606686</td>\n",
       "      <td>5.639268</td>\n",
       "      <td>6.330926</td>\n",
       "      <td>6.834661</td>\n",
       "      <td>6.725825</td>\n",
       "      <td>5.632979</td>\n",
       "      <td>6.230266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(+)E1A_r60_a104</th>\n",
       "      <td>2.894064</td>\n",
       "      <td>2.899291</td>\n",
       "      <td>2.815752</td>\n",
       "      <td>2.822751</td>\n",
       "      <td>3.035972</td>\n",
       "      <td>3.451708</td>\n",
       "      <td>3.365927</td>\n",
       "      <td>2.915322</td>\n",
       "      <td>3.178784</td>\n",
       "      <td>2.909366</td>\n",
       "      <td>...</td>\n",
       "      <td>2.813052</td>\n",
       "      <td>2.971318</td>\n",
       "      <td>3.223145</td>\n",
       "      <td>3.114489</td>\n",
       "      <td>3.113753</td>\n",
       "      <td>3.152049</td>\n",
       "      <td>3.279744</td>\n",
       "      <td>3.089397</td>\n",
       "      <td>2.978466</td>\n",
       "      <td>3.149935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(+)E1A_r60_a107</th>\n",
       "      <td>3.237023</td>\n",
       "      <td>3.233810</td>\n",
       "      <td>3.052772</td>\n",
       "      <td>2.913015</td>\n",
       "      <td>3.037258</td>\n",
       "      <td>2.997480</td>\n",
       "      <td>3.694422</td>\n",
       "      <td>2.851734</td>\n",
       "      <td>3.155538</td>\n",
       "      <td>3.100223</td>\n",
       "      <td>...</td>\n",
       "      <td>2.784098</td>\n",
       "      <td>2.810281</td>\n",
       "      <td>2.952121</td>\n",
       "      <td>2.872968</td>\n",
       "      <td>2.855768</td>\n",
       "      <td>2.886406</td>\n",
       "      <td>2.929406</td>\n",
       "      <td>2.928709</td>\n",
       "      <td>2.821016</td>\n",
       "      <td>2.908678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(+)E1A_r60_a135</th>\n",
       "      <td>2.954736</td>\n",
       "      <td>2.860286</td>\n",
       "      <td>2.936522</td>\n",
       "      <td>2.731988</td>\n",
       "      <td>2.946573</td>\n",
       "      <td>2.921638</td>\n",
       "      <td>3.183307</td>\n",
       "      <td>2.877388</td>\n",
       "      <td>2.897539</td>\n",
       "      <td>2.898822</td>\n",
       "      <td>...</td>\n",
       "      <td>3.015113</td>\n",
       "      <td>3.371869</td>\n",
       "      <td>3.247685</td>\n",
       "      <td>3.419323</td>\n",
       "      <td>3.670872</td>\n",
       "      <td>3.799645</td>\n",
       "      <td>4.004826</td>\n",
       "      <td>3.943948</td>\n",
       "      <td>3.313756</td>\n",
       "      <td>3.302009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mouse-PGK1_5</th>\n",
       "      <td>6.652913</td>\n",
       "      <td>6.739501</td>\n",
       "      <td>7.098698</td>\n",
       "      <td>7.341454</td>\n",
       "      <td>7.097640</td>\n",
       "      <td>7.305539</td>\n",
       "      <td>6.441029</td>\n",
       "      <td>6.780557</td>\n",
       "      <td>7.133135</td>\n",
       "      <td>7.222959</td>\n",
       "      <td>...</td>\n",
       "      <td>7.267308</td>\n",
       "      <td>6.833162</td>\n",
       "      <td>7.242861</td>\n",
       "      <td>7.123768</td>\n",
       "      <td>7.757611</td>\n",
       "      <td>7.613537</td>\n",
       "      <td>7.108137</td>\n",
       "      <td>7.041728</td>\n",
       "      <td>7.224681</td>\n",
       "      <td>6.686666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rat-GAPDH_3</th>\n",
       "      <td>9.665619</td>\n",
       "      <td>9.663037</td>\n",
       "      <td>9.832910</td>\n",
       "      <td>9.395596</td>\n",
       "      <td>9.562595</td>\n",
       "      <td>9.109386</td>\n",
       "      <td>9.993700</td>\n",
       "      <td>10.020462</td>\n",
       "      <td>9.721497</td>\n",
       "      <td>9.618459</td>\n",
       "      <td>...</td>\n",
       "      <td>9.284128</td>\n",
       "      <td>9.557268</td>\n",
       "      <td>9.594813</td>\n",
       "      <td>9.146444</td>\n",
       "      <td>9.382476</td>\n",
       "      <td>9.110078</td>\n",
       "      <td>9.928258</td>\n",
       "      <td>9.932738</td>\n",
       "      <td>8.526566</td>\n",
       "      <td>8.991116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rat-GAPDH_5</th>\n",
       "      <td>9.923443</td>\n",
       "      <td>10.188957</td>\n",
       "      <td>10.370885</td>\n",
       "      <td>10.922882</td>\n",
       "      <td>10.819334</td>\n",
       "      <td>10.613005</td>\n",
       "      <td>9.511038</td>\n",
       "      <td>9.847263</td>\n",
       "      <td>10.111858</td>\n",
       "      <td>10.107303</td>\n",
       "      <td>...</td>\n",
       "      <td>10.379840</td>\n",
       "      <td>9.993458</td>\n",
       "      <td>10.705560</td>\n",
       "      <td>11.044461</td>\n",
       "      <td>10.144645</td>\n",
       "      <td>9.949533</td>\n",
       "      <td>9.954214</td>\n",
       "      <td>9.842153</td>\n",
       "      <td>10.088426</td>\n",
       "      <td>10.082981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rat-PGK1_3</th>\n",
       "      <td>2.849316</td>\n",
       "      <td>3.204162</td>\n",
       "      <td>2.710797</td>\n",
       "      <td>2.904025</td>\n",
       "      <td>3.079639</td>\n",
       "      <td>3.073844</td>\n",
       "      <td>3.135625</td>\n",
       "      <td>3.499832</td>\n",
       "      <td>2.786003</td>\n",
       "      <td>2.844454</td>\n",
       "      <td>...</td>\n",
       "      <td>2.759463</td>\n",
       "      <td>3.272689</td>\n",
       "      <td>3.189556</td>\n",
       "      <td>2.902571</td>\n",
       "      <td>2.766141</td>\n",
       "      <td>2.941688</td>\n",
       "      <td>3.025670</td>\n",
       "      <td>2.826981</td>\n",
       "      <td>2.782573</td>\n",
       "      <td>2.770213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rat-PGK1_5</th>\n",
       "      <td>7.561878</td>\n",
       "      <td>7.559462</td>\n",
       "      <td>7.716728</td>\n",
       "      <td>8.033842</td>\n",
       "      <td>7.769111</td>\n",
       "      <td>8.077940</td>\n",
       "      <td>6.937684</td>\n",
       "      <td>7.147249</td>\n",
       "      <td>7.815745</td>\n",
       "      <td>7.903804</td>\n",
       "      <td>...</td>\n",
       "      <td>7.924271</td>\n",
       "      <td>7.438990</td>\n",
       "      <td>7.965310</td>\n",
       "      <td>7.802281</td>\n",
       "      <td>8.351095</td>\n",
       "      <td>8.314943</td>\n",
       "      <td>7.774011</td>\n",
       "      <td>7.770774</td>\n",
       "      <td>7.879103</td>\n",
       "      <td>7.347928</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111087 rows Ã— 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM2067403  GSM2067404  GSM2067405  GSM2067406  GSM2067407  \\\n",
       "ID                                                                            \n",
       "(+)E1A_r60_1      13.442002   13.263885   13.225509   13.144137   12.986026   \n",
       "(+)E1A_r60_3       5.764329    6.253843    6.890254    6.576528    6.647077   \n",
       "(+)E1A_r60_a104    2.894064    2.899291    2.815752    2.822751    3.035972   \n",
       "(+)E1A_r60_a107    3.237023    3.233810    3.052772    2.913015    3.037258   \n",
       "(+)E1A_r60_a135    2.954736    2.860286    2.936522    2.731988    2.946573   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "Mouse-PGK1_5       6.652913    6.739501    7.098698    7.341454    7.097640   \n",
       "Rat-GAPDH_3        9.665619    9.663037    9.832910    9.395596    9.562595   \n",
       "Rat-GAPDH_5        9.923443   10.188957   10.370885   10.922882   10.819334   \n",
       "Rat-PGK1_3         2.849316    3.204162    2.710797    2.904025    3.079639   \n",
       "Rat-PGK1_5         7.561878    7.559462    7.716728    8.033842    7.769111   \n",
       "\n",
       "                 GSM2067408  GSM2067409  GSM2067410  GSM2067411  GSM2067412  \\\n",
       "ID                                                                            \n",
       "(+)E1A_r60_1      13.849309   13.541306   14.152843   13.809382   14.053410   \n",
       "(+)E1A_r60_3       7.138355    7.172650    7.322565    6.515238    6.263576   \n",
       "(+)E1A_r60_a104    3.451708    3.365927    2.915322    3.178784    2.909366   \n",
       "(+)E1A_r60_a107    2.997480    3.694422    2.851734    3.155538    3.100223   \n",
       "(+)E1A_r60_a135    2.921638    3.183307    2.877388    2.897539    2.898822   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "Mouse-PGK1_5       7.305539    6.441029    6.780557    7.133135    7.222959   \n",
       "Rat-GAPDH_3        9.109386    9.993700   10.020462    9.721497    9.618459   \n",
       "Rat-GAPDH_5       10.613005    9.511038    9.847263   10.111858   10.107303   \n",
       "Rat-PGK1_3         3.073844    3.135625    3.499832    2.786003    2.844454   \n",
       "Rat-PGK1_5         8.077940    6.937684    7.147249    7.815745    7.903804   \n",
       "\n",
       "                 ...  GSM2067453  GSM2067454  GSM2067455  GSM2067456  \\\n",
       "ID               ...                                                   \n",
       "(+)E1A_r60_1     ...   13.750706   13.790617   13.543028   13.721816   \n",
       "(+)E1A_r60_3     ...    6.291416    6.574324    6.940987    6.606686   \n",
       "(+)E1A_r60_a104  ...    2.813052    2.971318    3.223145    3.114489   \n",
       "(+)E1A_r60_a107  ...    2.784098    2.810281    2.952121    2.872968   \n",
       "(+)E1A_r60_a135  ...    3.015113    3.371869    3.247685    3.419323   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "Mouse-PGK1_5     ...    7.267308    6.833162    7.242861    7.123768   \n",
       "Rat-GAPDH_3      ...    9.284128    9.557268    9.594813    9.146444   \n",
       "Rat-GAPDH_5      ...   10.379840    9.993458   10.705560   11.044461   \n",
       "Rat-PGK1_3       ...    2.759463    3.272689    3.189556    2.902571   \n",
       "Rat-PGK1_5       ...    7.924271    7.438990    7.965310    7.802281   \n",
       "\n",
       "                 GSM2067457  GSM2067458  GSM2067459  GSM2067460  GSM2067461  \\\n",
       "ID                                                                            \n",
       "(+)E1A_r60_1      14.025154   14.458712   14.075050   14.186631   14.150471   \n",
       "(+)E1A_r60_3       5.639268    6.330926    6.834661    6.725825    5.632979   \n",
       "(+)E1A_r60_a104    3.113753    3.152049    3.279744    3.089397    2.978466   \n",
       "(+)E1A_r60_a107    2.855768    2.886406    2.929406    2.928709    2.821016   \n",
       "(+)E1A_r60_a135    3.670872    3.799645    4.004826    3.943948    3.313756   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "Mouse-PGK1_5       7.757611    7.613537    7.108137    7.041728    7.224681   \n",
       "Rat-GAPDH_3        9.382476    9.110078    9.928258    9.932738    8.526566   \n",
       "Rat-GAPDH_5       10.144645    9.949533    9.954214    9.842153   10.088426   \n",
       "Rat-PGK1_3         2.766141    2.941688    3.025670    2.826981    2.782573   \n",
       "Rat-PGK1_5         8.351095    8.314943    7.774011    7.770774    7.879103   \n",
       "\n",
       "                 GSM2067462  \n",
       "ID                           \n",
       "(+)E1A_r60_1      13.071623  \n",
       "(+)E1A_r60_3       6.230266  \n",
       "(+)E1A_r60_a104    3.149935  \n",
       "(+)E1A_r60_a107    2.908678  \n",
       "(+)E1A_r60_a135    3.302009  \n",
       "...                     ...  \n",
       "Mouse-PGK1_5       6.686666  \n",
       "Rat-GAPDH_3        8.991116  \n",
       "Rat-GAPDH_5       10.082981  \n",
       "Rat-PGK1_3         2.770213  \n",
       "Rat-PGK1_5         7.347928  \n",
       "\n",
       "[111087 rows x 60 columns]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['A_19_P00315459', 'A_19_P00315492', 'A_19_P00315502', 'A_19_P00315506', 'A_19_P00315538'], 'CONTROL_TYPE': ['FALSE', 'FALSE', 'FALSE', 'FALSE', 'FALSE'], 'SEQUENCE': ['AGCCCCCACTGTTCCACTTATTGTGATGGTTTGTATATCTTTATTTCAAAGAAGATCTGT', 'AGGCAGCCTTGCTGTTGGGGGTTATTGGCAGCTGTTGGGGGTTAGAGACAGGACTCTCAT', 'AGCCGGGATCGGGTTGTTGTTAATTTCTTAAGCAATTTCTAAATTCTGTATTGACTCTCT', 'CAATGGATTCCATGTTTCTTTTTCTTGGGGGGAGCAGGGAGGGAGAAAGGTAGAAAAATG', 'CACAATGACCATCATTGAGGGCGATGTTTATGCTTCCATTGTTAGTTTAGATATTTTGTT'], 'TargetID': [nan, 'Q73P46', 'P01115', nan, nan], 'ncRNA_SeqID': [nan, nan, nan, nan, nan], 'Source': ['Agilent_humanG3V2', 'Agilent_humanG3V2', 'Agilent_humanG3V2', nan, nan], 'ncRNA_Accession': [nan, nan, nan, nan, nan], 'Chr': ['chrX', 'chr4', 'chr10', nan, nan], 'Start': [149131107.0, 129376376.0, 6780785.0, nan, nan], 'End': [149131166.0, 129376435.0, 6780844.0, nan, nan], 'strand': ['+', '+', '+', nan, nan], 'Description': [nan, 'Q73P46_TREDE (Q73P46) Branched-chain amino acid ABC transporter, permease protein, partial (5%) [THC2614189]', 'RASH_MSVHA (P01115) Transforming protein p29 precursor [Contains: Transforming protein p21], partial (6%) [THC2657193]', nan, nan], 'Genome': ['hg19', 'hg19', 'hg19', nan, nan], 'GeneSymbol': [nan, 'Q73P46', 'P01115', nan, nan], 'Seq_type': ['mRNA', 'mRNA', 'mRNA', nan, nan], 'ControlType': ['FALSE', 'FALSE', 'FALSE', nan, nan], 'EntrezGeneID': [nan, nan, nan, nan, nan], 'GenbankAccession': ['U66048', nan, nan, nan, nan], 'GeneName': [nan, nan, nan, nan, nan], 'Go': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan], 'UniGeneID': [nan, nan, nan, nan, nan], 'SPOT_ID': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['(+)E1A_r60_1', '(+)E1A_r60_3', '(+)E1A_r60_a104', '(+)E1A_r60_a107', '(+)E1A_r60_a135', '(+)E1A_r60_a20', '(+)E1A_r60_a22', '(+)E1A_r60_a97', '(+)E1A_r60_n11', '(+)E1A_r60_n9', '(-)3xSLv1', 'A_19_P00315459', 'A_19_P00315492', 'A_19_P00315502', 'A_19_P00315506', 'A_19_P00315538', 'A_19_P00315633', 'A_19_P00315668', 'A_19_P00315717', 'A_19_P00315718']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['A_19_P00315459', 'A_19_P00315492', 'A_19_P00315502', 'A_19_P00315506', 'A_19_P00315538'], 'CONTROL_TYPE': ['FALSE', 'FALSE', 'FALSE', 'FALSE', 'FALSE'], 'SEQUENCE': ['AGCCCCCACTGTTCCACTTATTGTGATGGTTTGTATATCTTTATTTCAAAGAAGATCTGT', 'AGGCAGCCTTGCTGTTGGGGGTTATTGGCAGCTGTTGGGGGTTAGAGACAGGACTCTCAT', 'AGCCGGGATCGGGTTGTTGTTAATTTCTTAAGCAATTTCTAAATTCTGTATTGACTCTCT', 'CAATGGATTCCATGTTTCTTTTTCTTGGGGGGAGCAGGGAGGGAGAAAGGTAGAAAAATG', 'CACAATGACCATCATTGAGGGCGATGTTTATGCTTCCATTGTTAGTTTAGATATTTTGTT'], 'TargetID': [nan, 'Q73P46', 'P01115', nan, nan], 'ncRNA_SeqID': [nan, nan, nan, nan, nan], 'Source': ['Agilent_humanG3V2', 'Agilent_humanG3V2', 'Agilent_humanG3V2', nan, nan], 'ncRNA_Accession': [nan, nan, nan, nan, nan], 'Chr': ['chrX', 'chr4', 'chr10', nan, nan], 'Start': [149131107.0, 129376376.0, 6780785.0, nan, nan], 'End': [149131166.0, 129376435.0, 6780844.0, nan, nan], 'strand': ['+', '+', '+', nan, nan], 'Description': [nan, 'Q73P46_TREDE (Q73P46) Branched-chain amino acid ABC transporter, permease protein, partial (5%) [THC2614189]', 'RASH_MSVHA (P01115) Transforming protein p29 precursor [Contains: Transforming protein p21], partial (6%) [THC2657193]', nan, nan], 'Genome': ['hg19', 'hg19', 'hg19', nan, nan], 'GeneSymbol': [nan, 'Q73P46', 'P01115', nan, nan], 'Seq_type': ['mRNA', 'mRNA', 'mRNA', nan, nan], 'ControlType': ['FALSE', 'FALSE', 'FALSE', nan, nan], 'EntrezGeneID': [nan, nan, nan, nan, nan], 'GenbankAccession': ['U66048', nan, nan, nan, nan], 'GeneName': [nan, nan, nan, nan, nan], 'Go': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan], 'UniGeneID': [nan, nan, nan, nan, nan], 'SPOT_ID': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'CONTROL_TYPE', 'SEQUENCE', 'TargetID', 'ncRNA_SeqID', 'Source',\n",
       "       'ncRNA_Accession', 'Chr', 'Start', 'End', 'strand', 'Description',\n",
       "       'Genome', 'GeneSymbol', 'Seq_type', 'ControlType', 'EntrezGeneID',\n",
       "       'GenbankAccession', 'GeneName', 'Go', 'GB_ACC', 'UniGeneID', 'SPOT_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GeneSymbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2067403</th>\n",
       "      <th>GSM2067404</th>\n",
       "      <th>GSM2067405</th>\n",
       "      <th>GSM2067406</th>\n",
       "      <th>GSM2067407</th>\n",
       "      <th>GSM2067408</th>\n",
       "      <th>GSM2067409</th>\n",
       "      <th>GSM2067410</th>\n",
       "      <th>GSM2067411</th>\n",
       "      <th>GSM2067412</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2067453</th>\n",
       "      <th>GSM2067454</th>\n",
       "      <th>GSM2067455</th>\n",
       "      <th>GSM2067456</th>\n",
       "      <th>GSM2067457</th>\n",
       "      <th>GSM2067458</th>\n",
       "      <th>GSM2067459</th>\n",
       "      <th>GSM2067460</th>\n",
       "      <th>GSM2067461</th>\n",
       "      <th>GSM2067462</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>6.714729</td>\n",
       "      <td>6.119238</td>\n",
       "      <td>6.588651</td>\n",
       "      <td>6.389377</td>\n",
       "      <td>6.489448</td>\n",
       "      <td>6.229496</td>\n",
       "      <td>5.642367</td>\n",
       "      <td>5.565992</td>\n",
       "      <td>6.170604</td>\n",
       "      <td>6.437280</td>\n",
       "      <td>...</td>\n",
       "      <td>6.436784</td>\n",
       "      <td>6.354268</td>\n",
       "      <td>6.692356</td>\n",
       "      <td>6.509643</td>\n",
       "      <td>7.310910</td>\n",
       "      <td>6.487669</td>\n",
       "      <td>6.933303</td>\n",
       "      <td>6.728533</td>\n",
       "      <td>6.522120</td>\n",
       "      <td>6.859007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>3.938725</td>\n",
       "      <td>3.857275</td>\n",
       "      <td>4.320075</td>\n",
       "      <td>4.589062</td>\n",
       "      <td>4.668073</td>\n",
       "      <td>4.644432</td>\n",
       "      <td>6.507755</td>\n",
       "      <td>5.445320</td>\n",
       "      <td>3.228480</td>\n",
       "      <td>4.336737</td>\n",
       "      <td>...</td>\n",
       "      <td>4.451320</td>\n",
       "      <td>4.461054</td>\n",
       "      <td>4.306510</td>\n",
       "      <td>3.974989</td>\n",
       "      <td>3.481818</td>\n",
       "      <td>4.010231</td>\n",
       "      <td>4.262363</td>\n",
       "      <td>4.172008</td>\n",
       "      <td>3.728118</td>\n",
       "      <td>3.784718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>4.105190</td>\n",
       "      <td>5.898689</td>\n",
       "      <td>5.163733</td>\n",
       "      <td>5.968357</td>\n",
       "      <td>4.846476</td>\n",
       "      <td>5.877680</td>\n",
       "      <td>5.523924</td>\n",
       "      <td>6.222440</td>\n",
       "      <td>6.007900</td>\n",
       "      <td>6.735513</td>\n",
       "      <td>...</td>\n",
       "      <td>6.042601</td>\n",
       "      <td>5.190834</td>\n",
       "      <td>5.677143</td>\n",
       "      <td>5.098562</td>\n",
       "      <td>5.875097</td>\n",
       "      <td>5.961306</td>\n",
       "      <td>5.425658</td>\n",
       "      <td>5.313376</td>\n",
       "      <td>6.113496</td>\n",
       "      <td>5.387186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>13.030699</td>\n",
       "      <td>12.865675</td>\n",
       "      <td>13.568920</td>\n",
       "      <td>12.454192</td>\n",
       "      <td>13.582437</td>\n",
       "      <td>12.652796</td>\n",
       "      <td>12.800940</td>\n",
       "      <td>11.776719</td>\n",
       "      <td>12.721724</td>\n",
       "      <td>12.280159</td>\n",
       "      <td>...</td>\n",
       "      <td>12.211297</td>\n",
       "      <td>11.890785</td>\n",
       "      <td>12.882605</td>\n",
       "      <td>12.581439</td>\n",
       "      <td>11.951515</td>\n",
       "      <td>11.085049</td>\n",
       "      <td>12.906422</td>\n",
       "      <td>12.599278</td>\n",
       "      <td>12.247601</td>\n",
       "      <td>12.555675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>2.704922</td>\n",
       "      <td>2.733084</td>\n",
       "      <td>3.909408</td>\n",
       "      <td>2.893169</td>\n",
       "      <td>2.692215</td>\n",
       "      <td>3.176106</td>\n",
       "      <td>3.038265</td>\n",
       "      <td>4.287126</td>\n",
       "      <td>2.768476</td>\n",
       "      <td>2.787102</td>\n",
       "      <td>...</td>\n",
       "      <td>3.411104</td>\n",
       "      <td>3.899433</td>\n",
       "      <td>4.059085</td>\n",
       "      <td>2.905740</td>\n",
       "      <td>2.999079</td>\n",
       "      <td>3.518736</td>\n",
       "      <td>3.275482</td>\n",
       "      <td>2.708914</td>\n",
       "      <td>2.969698</td>\n",
       "      <td>3.287521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>6.206871</td>\n",
       "      <td>6.407975</td>\n",
       "      <td>6.309663</td>\n",
       "      <td>6.479846</td>\n",
       "      <td>6.401652</td>\n",
       "      <td>6.667216</td>\n",
       "      <td>6.080510</td>\n",
       "      <td>5.717164</td>\n",
       "      <td>6.664669</td>\n",
       "      <td>5.979232</td>\n",
       "      <td>...</td>\n",
       "      <td>6.647157</td>\n",
       "      <td>6.405932</td>\n",
       "      <td>6.357309</td>\n",
       "      <td>6.456726</td>\n",
       "      <td>6.496564</td>\n",
       "      <td>6.893305</td>\n",
       "      <td>6.206098</td>\n",
       "      <td>6.251587</td>\n",
       "      <td>6.796412</td>\n",
       "      <td>6.615562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>3.020767</td>\n",
       "      <td>3.026273</td>\n",
       "      <td>3.113522</td>\n",
       "      <td>2.773175</td>\n",
       "      <td>2.981554</td>\n",
       "      <td>3.513956</td>\n",
       "      <td>3.391591</td>\n",
       "      <td>3.025322</td>\n",
       "      <td>3.290877</td>\n",
       "      <td>3.548486</td>\n",
       "      <td>...</td>\n",
       "      <td>2.735566</td>\n",
       "      <td>2.748930</td>\n",
       "      <td>2.876586</td>\n",
       "      <td>2.828128</td>\n",
       "      <td>2.962675</td>\n",
       "      <td>2.874528</td>\n",
       "      <td>2.942508</td>\n",
       "      <td>3.252552</td>\n",
       "      <td>2.821534</td>\n",
       "      <td>3.412952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>5.951725</td>\n",
       "      <td>6.141165</td>\n",
       "      <td>5.639593</td>\n",
       "      <td>6.538147</td>\n",
       "      <td>5.849266</td>\n",
       "      <td>7.026554</td>\n",
       "      <td>5.296118</td>\n",
       "      <td>5.633491</td>\n",
       "      <td>6.056791</td>\n",
       "      <td>6.068389</td>\n",
       "      <td>...</td>\n",
       "      <td>6.347763</td>\n",
       "      <td>6.040958</td>\n",
       "      <td>6.040609</td>\n",
       "      <td>6.360076</td>\n",
       "      <td>6.197123</td>\n",
       "      <td>5.809800</td>\n",
       "      <td>5.440079</td>\n",
       "      <td>5.502089</td>\n",
       "      <td>6.425600</td>\n",
       "      <td>6.460133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>12.271689</td>\n",
       "      <td>12.290836</td>\n",
       "      <td>12.674829</td>\n",
       "      <td>13.083159</td>\n",
       "      <td>12.931067</td>\n",
       "      <td>12.483398</td>\n",
       "      <td>12.510457</td>\n",
       "      <td>12.603576</td>\n",
       "      <td>13.032246</td>\n",
       "      <td>13.294943</td>\n",
       "      <td>...</td>\n",
       "      <td>13.338604</td>\n",
       "      <td>12.802153</td>\n",
       "      <td>12.110378</td>\n",
       "      <td>12.578838</td>\n",
       "      <td>12.872514</td>\n",
       "      <td>12.954499</td>\n",
       "      <td>12.288324</td>\n",
       "      <td>12.264850</td>\n",
       "      <td>13.421458</td>\n",
       "      <td>12.969342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>3.938076</td>\n",
       "      <td>4.905338</td>\n",
       "      <td>4.189789</td>\n",
       "      <td>5.280000</td>\n",
       "      <td>4.669816</td>\n",
       "      <td>5.285511</td>\n",
       "      <td>5.079488</td>\n",
       "      <td>4.814769</td>\n",
       "      <td>4.740775</td>\n",
       "      <td>4.656932</td>\n",
       "      <td>...</td>\n",
       "      <td>5.386587</td>\n",
       "      <td>5.418139</td>\n",
       "      <td>4.437177</td>\n",
       "      <td>5.433254</td>\n",
       "      <td>5.885852</td>\n",
       "      <td>4.147930</td>\n",
       "      <td>4.563690</td>\n",
       "      <td>4.904048</td>\n",
       "      <td>5.616683</td>\n",
       "      <td>5.282989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14029 rows Ã— 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM2067403  GSM2067404  GSM2067405  GSM2067406  GSM2067407  \\\n",
       "A1BG      6.714729    6.119238    6.588651    6.389377    6.489448   \n",
       "A1CF      3.938725    3.857275    4.320075    4.589062    4.668073   \n",
       "A2M       4.105190    5.898689    5.163733    5.968357    4.846476   \n",
       "A4GALT   13.030699   12.865675   13.568920   12.454192   13.582437   \n",
       "A4GNT     2.704922    2.733084    3.909408    2.893169    2.692215   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZXDC      6.206871    6.407975    6.309663    6.479846    6.401652   \n",
       "ZYG11A    3.020767    3.026273    3.113522    2.773175    2.981554   \n",
       "ZYG11B    5.951725    6.141165    5.639593    6.538147    5.849266   \n",
       "ZYX      12.271689   12.290836   12.674829   13.083159   12.931067   \n",
       "ZZZ3      3.938076    4.905338    4.189789    5.280000    4.669816   \n",
       "\n",
       "        GSM2067408  GSM2067409  GSM2067410  GSM2067411  GSM2067412  ...  \\\n",
       "A1BG      6.229496    5.642367    5.565992    6.170604    6.437280  ...   \n",
       "A1CF      4.644432    6.507755    5.445320    3.228480    4.336737  ...   \n",
       "A2M       5.877680    5.523924    6.222440    6.007900    6.735513  ...   \n",
       "A4GALT   12.652796   12.800940   11.776719   12.721724   12.280159  ...   \n",
       "A4GNT     3.176106    3.038265    4.287126    2.768476    2.787102  ...   \n",
       "...            ...         ...         ...         ...         ...  ...   \n",
       "ZXDC      6.667216    6.080510    5.717164    6.664669    5.979232  ...   \n",
       "ZYG11A    3.513956    3.391591    3.025322    3.290877    3.548486  ...   \n",
       "ZYG11B    7.026554    5.296118    5.633491    6.056791    6.068389  ...   \n",
       "ZYX      12.483398   12.510457   12.603576   13.032246   13.294943  ...   \n",
       "ZZZ3      5.285511    5.079488    4.814769    4.740775    4.656932  ...   \n",
       "\n",
       "        GSM2067453  GSM2067454  GSM2067455  GSM2067456  GSM2067457  \\\n",
       "A1BG      6.436784    6.354268    6.692356    6.509643    7.310910   \n",
       "A1CF      4.451320    4.461054    4.306510    3.974989    3.481818   \n",
       "A2M       6.042601    5.190834    5.677143    5.098562    5.875097   \n",
       "A4GALT   12.211297   11.890785   12.882605   12.581439   11.951515   \n",
       "A4GNT     3.411104    3.899433    4.059085    2.905740    2.999079   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZXDC      6.647157    6.405932    6.357309    6.456726    6.496564   \n",
       "ZYG11A    2.735566    2.748930    2.876586    2.828128    2.962675   \n",
       "ZYG11B    6.347763    6.040958    6.040609    6.360076    6.197123   \n",
       "ZYX      13.338604   12.802153   12.110378   12.578838   12.872514   \n",
       "ZZZ3      5.386587    5.418139    4.437177    5.433254    5.885852   \n",
       "\n",
       "        GSM2067458  GSM2067459  GSM2067460  GSM2067461  GSM2067462  \n",
       "A1BG      6.487669    6.933303    6.728533    6.522120    6.859007  \n",
       "A1CF      4.010231    4.262363    4.172008    3.728118    3.784718  \n",
       "A2M       5.961306    5.425658    5.313376    6.113496    5.387186  \n",
       "A4GALT   11.085049   12.906422   12.599278   12.247601   12.555675  \n",
       "A4GNT     3.518736    3.275482    2.708914    2.969698    3.287521  \n",
       "...            ...         ...         ...         ...         ...  \n",
       "ZXDC      6.893305    6.206098    6.251587    6.796412    6.615562  \n",
       "ZYG11A    2.874528    2.942508    3.252552    2.821534    3.412952  \n",
       "ZYG11B    5.809800    5.440079    5.502089    6.425600    6.460133  \n",
       "ZYX      12.954499   12.288324   12.264850   13.421458   12.969342  \n",
       "ZZZ3      4.147930    4.563690    4.904048    5.616683    5.282989  \n",
       "\n",
       "[14029 rows x 60 columns]"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2067403</th>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.714729</td>\n",
       "      <td>3.938725</td>\n",
       "      <td>4.105190</td>\n",
       "      <td>13.030699</td>\n",
       "      <td>2.704922</td>\n",
       "      <td>11.219065</td>\n",
       "      <td>6.193231</td>\n",
       "      <td>...</td>\n",
       "      <td>8.472954</td>\n",
       "      <td>4.146875</td>\n",
       "      <td>5.884128</td>\n",
       "      <td>3.879921</td>\n",
       "      <td>7.125728</td>\n",
       "      <td>6.206871</td>\n",
       "      <td>3.020767</td>\n",
       "      <td>5.951725</td>\n",
       "      <td>12.271689</td>\n",
       "      <td>3.938076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067404</th>\n",
       "      <td>1.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.119238</td>\n",
       "      <td>3.857275</td>\n",
       "      <td>5.898689</td>\n",
       "      <td>12.865675</td>\n",
       "      <td>2.733084</td>\n",
       "      <td>10.718920</td>\n",
       "      <td>5.529508</td>\n",
       "      <td>...</td>\n",
       "      <td>8.837225</td>\n",
       "      <td>5.017093</td>\n",
       "      <td>6.116274</td>\n",
       "      <td>4.135404</td>\n",
       "      <td>7.121595</td>\n",
       "      <td>6.407975</td>\n",
       "      <td>3.026273</td>\n",
       "      <td>6.141165</td>\n",
       "      <td>12.290836</td>\n",
       "      <td>4.905338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067405</th>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.588651</td>\n",
       "      <td>4.320075</td>\n",
       "      <td>5.163733</td>\n",
       "      <td>13.568920</td>\n",
       "      <td>3.909408</td>\n",
       "      <td>11.084585</td>\n",
       "      <td>6.278926</td>\n",
       "      <td>...</td>\n",
       "      <td>8.916005</td>\n",
       "      <td>5.166640</td>\n",
       "      <td>5.942174</td>\n",
       "      <td>2.728597</td>\n",
       "      <td>7.344893</td>\n",
       "      <td>6.309663</td>\n",
       "      <td>3.113522</td>\n",
       "      <td>5.639593</td>\n",
       "      <td>12.674829</td>\n",
       "      <td>4.189789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067406</th>\n",
       "      <td>1.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.389377</td>\n",
       "      <td>4.589062</td>\n",
       "      <td>5.968357</td>\n",
       "      <td>12.454192</td>\n",
       "      <td>2.893169</td>\n",
       "      <td>10.745497</td>\n",
       "      <td>5.896101</td>\n",
       "      <td>...</td>\n",
       "      <td>8.870725</td>\n",
       "      <td>4.644601</td>\n",
       "      <td>5.981628</td>\n",
       "      <td>4.986273</td>\n",
       "      <td>7.135086</td>\n",
       "      <td>6.479846</td>\n",
       "      <td>2.773175</td>\n",
       "      <td>6.538147</td>\n",
       "      <td>13.083159</td>\n",
       "      <td>5.280000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067407</th>\n",
       "      <td>1.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.489448</td>\n",
       "      <td>4.668073</td>\n",
       "      <td>4.846476</td>\n",
       "      <td>13.582437</td>\n",
       "      <td>2.692215</td>\n",
       "      <td>10.799071</td>\n",
       "      <td>6.265001</td>\n",
       "      <td>...</td>\n",
       "      <td>8.773541</td>\n",
       "      <td>3.564052</td>\n",
       "      <td>6.142061</td>\n",
       "      <td>4.558930</td>\n",
       "      <td>7.214107</td>\n",
       "      <td>6.401652</td>\n",
       "      <td>2.981554</td>\n",
       "      <td>5.849266</td>\n",
       "      <td>12.931067</td>\n",
       "      <td>4.669816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067408</th>\n",
       "      <td>1.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.229496</td>\n",
       "      <td>4.644432</td>\n",
       "      <td>5.877680</td>\n",
       "      <td>12.652796</td>\n",
       "      <td>3.176106</td>\n",
       "      <td>9.883424</td>\n",
       "      <td>6.285170</td>\n",
       "      <td>...</td>\n",
       "      <td>8.755016</td>\n",
       "      <td>5.112708</td>\n",
       "      <td>6.385875</td>\n",
       "      <td>4.676870</td>\n",
       "      <td>6.742880</td>\n",
       "      <td>6.667216</td>\n",
       "      <td>3.513956</td>\n",
       "      <td>7.026554</td>\n",
       "      <td>12.483398</td>\n",
       "      <td>5.285511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067409</th>\n",
       "      <td>1.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.642367</td>\n",
       "      <td>6.507755</td>\n",
       "      <td>5.523924</td>\n",
       "      <td>12.800940</td>\n",
       "      <td>3.038265</td>\n",
       "      <td>10.526562</td>\n",
       "      <td>6.054632</td>\n",
       "      <td>...</td>\n",
       "      <td>8.459496</td>\n",
       "      <td>5.665781</td>\n",
       "      <td>6.182136</td>\n",
       "      <td>4.099351</td>\n",
       "      <td>6.705152</td>\n",
       "      <td>6.080510</td>\n",
       "      <td>3.391591</td>\n",
       "      <td>5.296118</td>\n",
       "      <td>12.510457</td>\n",
       "      <td>5.079488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067410</th>\n",
       "      <td>1.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.565992</td>\n",
       "      <td>5.445320</td>\n",
       "      <td>6.222440</td>\n",
       "      <td>11.776719</td>\n",
       "      <td>4.287126</td>\n",
       "      <td>10.491643</td>\n",
       "      <td>6.232655</td>\n",
       "      <td>...</td>\n",
       "      <td>8.678071</td>\n",
       "      <td>4.061905</td>\n",
       "      <td>5.991161</td>\n",
       "      <td>4.052005</td>\n",
       "      <td>7.071524</td>\n",
       "      <td>5.717164</td>\n",
       "      <td>3.025322</td>\n",
       "      <td>5.633491</td>\n",
       "      <td>12.603576</td>\n",
       "      <td>4.814769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067411</th>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.170604</td>\n",
       "      <td>3.228480</td>\n",
       "      <td>6.007900</td>\n",
       "      <td>12.721724</td>\n",
       "      <td>2.768476</td>\n",
       "      <td>11.095754</td>\n",
       "      <td>6.630271</td>\n",
       "      <td>...</td>\n",
       "      <td>8.630165</td>\n",
       "      <td>4.591509</td>\n",
       "      <td>6.099095</td>\n",
       "      <td>3.889528</td>\n",
       "      <td>6.957309</td>\n",
       "      <td>6.664669</td>\n",
       "      <td>3.290877</td>\n",
       "      <td>6.056791</td>\n",
       "      <td>13.032246</td>\n",
       "      <td>4.740775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067412</th>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.437280</td>\n",
       "      <td>4.336737</td>\n",
       "      <td>6.735513</td>\n",
       "      <td>12.280159</td>\n",
       "      <td>2.787102</td>\n",
       "      <td>10.937666</td>\n",
       "      <td>6.262236</td>\n",
       "      <td>...</td>\n",
       "      <td>8.848257</td>\n",
       "      <td>4.779469</td>\n",
       "      <td>6.101059</td>\n",
       "      <td>4.707293</td>\n",
       "      <td>7.179975</td>\n",
       "      <td>5.979232</td>\n",
       "      <td>3.548486</td>\n",
       "      <td>6.068389</td>\n",
       "      <td>13.294943</td>\n",
       "      <td>4.656932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067413</th>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.624097</td>\n",
       "      <td>4.086302</td>\n",
       "      <td>5.618996</td>\n",
       "      <td>13.730714</td>\n",
       "      <td>3.903112</td>\n",
       "      <td>10.699893</td>\n",
       "      <td>6.376582</td>\n",
       "      <td>...</td>\n",
       "      <td>8.572907</td>\n",
       "      <td>5.004915</td>\n",
       "      <td>5.937257</td>\n",
       "      <td>5.041395</td>\n",
       "      <td>6.972745</td>\n",
       "      <td>6.073244</td>\n",
       "      <td>2.817874</td>\n",
       "      <td>6.416759</td>\n",
       "      <td>12.373903</td>\n",
       "      <td>5.321508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067414</th>\n",
       "      <td>1.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.508125</td>\n",
       "      <td>4.479011</td>\n",
       "      <td>5.697313</td>\n",
       "      <td>13.405120</td>\n",
       "      <td>3.854834</td>\n",
       "      <td>10.876848</td>\n",
       "      <td>6.166716</td>\n",
       "      <td>...</td>\n",
       "      <td>8.586086</td>\n",
       "      <td>3.447876</td>\n",
       "      <td>6.369736</td>\n",
       "      <td>3.533126</td>\n",
       "      <td>7.343472</td>\n",
       "      <td>6.380209</td>\n",
       "      <td>2.848518</td>\n",
       "      <td>5.788387</td>\n",
       "      <td>12.820009</td>\n",
       "      <td>4.926111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067415</th>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.395001</td>\n",
       "      <td>3.703604</td>\n",
       "      <td>5.590061</td>\n",
       "      <td>12.015218</td>\n",
       "      <td>3.254557</td>\n",
       "      <td>10.477460</td>\n",
       "      <td>6.270210</td>\n",
       "      <td>...</td>\n",
       "      <td>8.996468</td>\n",
       "      <td>3.963160</td>\n",
       "      <td>5.912210</td>\n",
       "      <td>4.288434</td>\n",
       "      <td>6.621121</td>\n",
       "      <td>6.607335</td>\n",
       "      <td>3.140500</td>\n",
       "      <td>6.342705</td>\n",
       "      <td>12.940755</td>\n",
       "      <td>4.109571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067416</th>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.398784</td>\n",
       "      <td>4.504799</td>\n",
       "      <td>5.234035</td>\n",
       "      <td>12.138115</td>\n",
       "      <td>3.756422</td>\n",
       "      <td>10.797807</td>\n",
       "      <td>6.289853</td>\n",
       "      <td>...</td>\n",
       "      <td>8.771525</td>\n",
       "      <td>4.398438</td>\n",
       "      <td>6.232646</td>\n",
       "      <td>3.979691</td>\n",
       "      <td>6.619864</td>\n",
       "      <td>6.575380</td>\n",
       "      <td>3.098263</td>\n",
       "      <td>5.728942</td>\n",
       "      <td>13.071623</td>\n",
       "      <td>4.070495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067417</th>\n",
       "      <td>1.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.955859</td>\n",
       "      <td>3.523432</td>\n",
       "      <td>6.458198</td>\n",
       "      <td>12.381153</td>\n",
       "      <td>3.165949</td>\n",
       "      <td>11.393041</td>\n",
       "      <td>6.272181</td>\n",
       "      <td>...</td>\n",
       "      <td>8.896212</td>\n",
       "      <td>5.605754</td>\n",
       "      <td>6.731539</td>\n",
       "      <td>5.507618</td>\n",
       "      <td>7.359767</td>\n",
       "      <td>7.044290</td>\n",
       "      <td>3.271596</td>\n",
       "      <td>6.527856</td>\n",
       "      <td>13.238666</td>\n",
       "      <td>5.625271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067418</th>\n",
       "      <td>1.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.926815</td>\n",
       "      <td>2.940408</td>\n",
       "      <td>6.273612</td>\n",
       "      <td>12.155148</td>\n",
       "      <td>3.010371</td>\n",
       "      <td>11.172894</td>\n",
       "      <td>6.242284</td>\n",
       "      <td>...</td>\n",
       "      <td>9.155986</td>\n",
       "      <td>5.731468</td>\n",
       "      <td>6.437933</td>\n",
       "      <td>5.371328</td>\n",
       "      <td>7.483693</td>\n",
       "      <td>7.304184</td>\n",
       "      <td>2.870487</td>\n",
       "      <td>7.111848</td>\n",
       "      <td>13.220077</td>\n",
       "      <td>5.601524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067419</th>\n",
       "      <td>1.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.700031</td>\n",
       "      <td>3.603868</td>\n",
       "      <td>6.608646</td>\n",
       "      <td>13.373009</td>\n",
       "      <td>3.368898</td>\n",
       "      <td>11.111868</td>\n",
       "      <td>6.140389</td>\n",
       "      <td>...</td>\n",
       "      <td>8.790861</td>\n",
       "      <td>4.912009</td>\n",
       "      <td>6.122825</td>\n",
       "      <td>4.528871</td>\n",
       "      <td>7.217832</td>\n",
       "      <td>6.580385</td>\n",
       "      <td>2.890108</td>\n",
       "      <td>6.398405</td>\n",
       "      <td>12.554396</td>\n",
       "      <td>5.229794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067420</th>\n",
       "      <td>1.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.420770</td>\n",
       "      <td>4.308397</td>\n",
       "      <td>4.419219</td>\n",
       "      <td>12.657779</td>\n",
       "      <td>4.182192</td>\n",
       "      <td>10.265317</td>\n",
       "      <td>5.575451</td>\n",
       "      <td>...</td>\n",
       "      <td>8.938309</td>\n",
       "      <td>4.391050</td>\n",
       "      <td>5.829993</td>\n",
       "      <td>3.814391</td>\n",
       "      <td>6.844831</td>\n",
       "      <td>6.732603</td>\n",
       "      <td>2.956524</td>\n",
       "      <td>7.408227</td>\n",
       "      <td>12.217965</td>\n",
       "      <td>4.816319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067421</th>\n",
       "      <td>1.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.865487</td>\n",
       "      <td>3.445940</td>\n",
       "      <td>5.425325</td>\n",
       "      <td>12.538816</td>\n",
       "      <td>3.530409</td>\n",
       "      <td>10.940994</td>\n",
       "      <td>6.028025</td>\n",
       "      <td>...</td>\n",
       "      <td>8.903367</td>\n",
       "      <td>5.238989</td>\n",
       "      <td>6.509625</td>\n",
       "      <td>5.306362</td>\n",
       "      <td>7.448940</td>\n",
       "      <td>7.041186</td>\n",
       "      <td>2.953032</td>\n",
       "      <td>6.538046</td>\n",
       "      <td>12.637005</td>\n",
       "      <td>5.844087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067422</th>\n",
       "      <td>1.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.839920</td>\n",
       "      <td>3.690498</td>\n",
       "      <td>6.696085</td>\n",
       "      <td>12.435116</td>\n",
       "      <td>3.284214</td>\n",
       "      <td>11.218464</td>\n",
       "      <td>6.190138</td>\n",
       "      <td>...</td>\n",
       "      <td>8.994142</td>\n",
       "      <td>5.549764</td>\n",
       "      <td>6.738438</td>\n",
       "      <td>5.249587</td>\n",
       "      <td>7.580654</td>\n",
       "      <td>6.882779</td>\n",
       "      <td>2.926269</td>\n",
       "      <td>6.844660</td>\n",
       "      <td>13.024989</td>\n",
       "      <td>5.728438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067423</th>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.770218</td>\n",
       "      <td>4.635688</td>\n",
       "      <td>5.216162</td>\n",
       "      <td>12.126523</td>\n",
       "      <td>4.427112</td>\n",
       "      <td>10.802723</td>\n",
       "      <td>6.136266</td>\n",
       "      <td>...</td>\n",
       "      <td>8.627634</td>\n",
       "      <td>4.336113</td>\n",
       "      <td>5.497256</td>\n",
       "      <td>4.016527</td>\n",
       "      <td>7.139729</td>\n",
       "      <td>6.044130</td>\n",
       "      <td>2.774483</td>\n",
       "      <td>5.807989</td>\n",
       "      <td>12.337760</td>\n",
       "      <td>4.647233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067424</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.359819</td>\n",
       "      <td>4.801514</td>\n",
       "      <td>4.932057</td>\n",
       "      <td>12.145253</td>\n",
       "      <td>3.818097</td>\n",
       "      <td>10.907265</td>\n",
       "      <td>6.121207</td>\n",
       "      <td>...</td>\n",
       "      <td>8.987240</td>\n",
       "      <td>4.889013</td>\n",
       "      <td>6.417872</td>\n",
       "      <td>4.120697</td>\n",
       "      <td>6.983198</td>\n",
       "      <td>6.486918</td>\n",
       "      <td>2.733062</td>\n",
       "      <td>6.645017</td>\n",
       "      <td>12.957374</td>\n",
       "      <td>4.746378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067425</th>\n",
       "      <td>1.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.663395</td>\n",
       "      <td>3.894671</td>\n",
       "      <td>5.955735</td>\n",
       "      <td>13.277133</td>\n",
       "      <td>2.958530</td>\n",
       "      <td>10.570036</td>\n",
       "      <td>5.945974</td>\n",
       "      <td>...</td>\n",
       "      <td>8.630093</td>\n",
       "      <td>4.538773</td>\n",
       "      <td>6.209540</td>\n",
       "      <td>4.679893</td>\n",
       "      <td>7.238637</td>\n",
       "      <td>6.523522</td>\n",
       "      <td>3.026820</td>\n",
       "      <td>6.441123</td>\n",
       "      <td>12.399323</td>\n",
       "      <td>5.098562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067426</th>\n",
       "      <td>1.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.663239</td>\n",
       "      <td>4.511771</td>\n",
       "      <td>5.806159</td>\n",
       "      <td>12.673718</td>\n",
       "      <td>3.400782</td>\n",
       "      <td>10.956592</td>\n",
       "      <td>6.158459</td>\n",
       "      <td>...</td>\n",
       "      <td>8.758475</td>\n",
       "      <td>4.394046</td>\n",
       "      <td>6.324679</td>\n",
       "      <td>4.530297</td>\n",
       "      <td>7.257558</td>\n",
       "      <td>6.820784</td>\n",
       "      <td>2.899090</td>\n",
       "      <td>6.546835</td>\n",
       "      <td>12.772914</td>\n",
       "      <td>5.087595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067427</th>\n",
       "      <td>1.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.542545</td>\n",
       "      <td>4.366632</td>\n",
       "      <td>5.442149</td>\n",
       "      <td>12.275525</td>\n",
       "      <td>2.735066</td>\n",
       "      <td>11.355851</td>\n",
       "      <td>6.237150</td>\n",
       "      <td>...</td>\n",
       "      <td>9.033045</td>\n",
       "      <td>3.229970</td>\n",
       "      <td>5.884744</td>\n",
       "      <td>4.018757</td>\n",
       "      <td>7.167019</td>\n",
       "      <td>6.326521</td>\n",
       "      <td>3.125467</td>\n",
       "      <td>6.801677</td>\n",
       "      <td>13.077705</td>\n",
       "      <td>4.810358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067428</th>\n",
       "      <td>1.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.877935</td>\n",
       "      <td>3.891802</td>\n",
       "      <td>4.252012</td>\n",
       "      <td>11.616556</td>\n",
       "      <td>3.679624</td>\n",
       "      <td>10.997719</td>\n",
       "      <td>5.853625</td>\n",
       "      <td>...</td>\n",
       "      <td>9.273019</td>\n",
       "      <td>5.005182</td>\n",
       "      <td>5.920124</td>\n",
       "      <td>4.349660</td>\n",
       "      <td>6.814784</td>\n",
       "      <td>7.202935</td>\n",
       "      <td>3.026861</td>\n",
       "      <td>6.738949</td>\n",
       "      <td>12.957374</td>\n",
       "      <td>5.203513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067429</th>\n",
       "      <td>1.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.517502</td>\n",
       "      <td>4.347663</td>\n",
       "      <td>5.699042</td>\n",
       "      <td>12.834258</td>\n",
       "      <td>3.219715</td>\n",
       "      <td>10.688426</td>\n",
       "      <td>5.963651</td>\n",
       "      <td>...</td>\n",
       "      <td>8.611595</td>\n",
       "      <td>4.199718</td>\n",
       "      <td>5.700428</td>\n",
       "      <td>4.972825</td>\n",
       "      <td>7.023918</td>\n",
       "      <td>6.558665</td>\n",
       "      <td>2.869833</td>\n",
       "      <td>6.466369</td>\n",
       "      <td>11.803370</td>\n",
       "      <td>5.039581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067430</th>\n",
       "      <td>1.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.100269</td>\n",
       "      <td>4.299640</td>\n",
       "      <td>4.012525</td>\n",
       "      <td>13.097274</td>\n",
       "      <td>3.685450</td>\n",
       "      <td>10.906361</td>\n",
       "      <td>6.371967</td>\n",
       "      <td>...</td>\n",
       "      <td>8.248276</td>\n",
       "      <td>3.136578</td>\n",
       "      <td>5.498039</td>\n",
       "      <td>4.274403</td>\n",
       "      <td>7.583811</td>\n",
       "      <td>5.958409</td>\n",
       "      <td>3.068320</td>\n",
       "      <td>5.905375</td>\n",
       "      <td>11.859551</td>\n",
       "      <td>5.343675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067431</th>\n",
       "      <td>1.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.467290</td>\n",
       "      <td>4.048092</td>\n",
       "      <td>5.393207</td>\n",
       "      <td>13.331580</td>\n",
       "      <td>3.237607</td>\n",
       "      <td>9.928258</td>\n",
       "      <td>5.895905</td>\n",
       "      <td>...</td>\n",
       "      <td>9.171186</td>\n",
       "      <td>5.097364</td>\n",
       "      <td>6.366438</td>\n",
       "      <td>6.166522</td>\n",
       "      <td>6.801190</td>\n",
       "      <td>7.086836</td>\n",
       "      <td>2.805352</td>\n",
       "      <td>7.007291</td>\n",
       "      <td>13.104760</td>\n",
       "      <td>5.478596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067432</th>\n",
       "      <td>1.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.671121</td>\n",
       "      <td>4.381368</td>\n",
       "      <td>6.989244</td>\n",
       "      <td>12.360163</td>\n",
       "      <td>3.558934</td>\n",
       "      <td>11.041781</td>\n",
       "      <td>6.190573</td>\n",
       "      <td>...</td>\n",
       "      <td>8.743233</td>\n",
       "      <td>4.372523</td>\n",
       "      <td>5.765166</td>\n",
       "      <td>4.768969</td>\n",
       "      <td>7.329707</td>\n",
       "      <td>6.971662</td>\n",
       "      <td>2.899214</td>\n",
       "      <td>6.184332</td>\n",
       "      <td>13.217264</td>\n",
       "      <td>4.937017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067433</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.650069</td>\n",
       "      <td>3.928323</td>\n",
       "      <td>5.469918</td>\n",
       "      <td>13.393532</td>\n",
       "      <td>2.809327</td>\n",
       "      <td>10.821529</td>\n",
       "      <td>6.230222</td>\n",
       "      <td>...</td>\n",
       "      <td>8.686610</td>\n",
       "      <td>5.212686</td>\n",
       "      <td>6.449946</td>\n",
       "      <td>5.164700</td>\n",
       "      <td>6.903481</td>\n",
       "      <td>6.541790</td>\n",
       "      <td>3.213683</td>\n",
       "      <td>6.583720</td>\n",
       "      <td>12.431108</td>\n",
       "      <td>5.720720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067434</th>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.990864</td>\n",
       "      <td>4.462671</td>\n",
       "      <td>6.327537</td>\n",
       "      <td>13.056033</td>\n",
       "      <td>2.787889</td>\n",
       "      <td>10.704703</td>\n",
       "      <td>5.867151</td>\n",
       "      <td>...</td>\n",
       "      <td>8.755357</td>\n",
       "      <td>4.877190</td>\n",
       "      <td>6.319072</td>\n",
       "      <td>4.974823</td>\n",
       "      <td>7.240470</td>\n",
       "      <td>5.733144</td>\n",
       "      <td>3.366409</td>\n",
       "      <td>6.213997</td>\n",
       "      <td>12.211297</td>\n",
       "      <td>5.271974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067435</th>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.717298</td>\n",
       "      <td>3.816540</td>\n",
       "      <td>6.621448</td>\n",
       "      <td>13.331580</td>\n",
       "      <td>3.488667</td>\n",
       "      <td>10.716845</td>\n",
       "      <td>5.750632</td>\n",
       "      <td>...</td>\n",
       "      <td>8.492819</td>\n",
       "      <td>5.080455</td>\n",
       "      <td>6.644737</td>\n",
       "      <td>4.943316</td>\n",
       "      <td>7.308149</td>\n",
       "      <td>6.303049</td>\n",
       "      <td>2.803965</td>\n",
       "      <td>6.304306</td>\n",
       "      <td>12.129213</td>\n",
       "      <td>5.658377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067436</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.055410</td>\n",
       "      <td>5.362066</td>\n",
       "      <td>3.588085</td>\n",
       "      <td>12.241908</td>\n",
       "      <td>3.464015</td>\n",
       "      <td>10.516702</td>\n",
       "      <td>6.380311</td>\n",
       "      <td>...</td>\n",
       "      <td>8.438438</td>\n",
       "      <td>4.235894</td>\n",
       "      <td>5.786593</td>\n",
       "      <td>4.708460</td>\n",
       "      <td>6.783018</td>\n",
       "      <td>5.907134</td>\n",
       "      <td>2.986399</td>\n",
       "      <td>6.238057</td>\n",
       "      <td>12.691983</td>\n",
       "      <td>4.899414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067437</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.788504</td>\n",
       "      <td>4.898884</td>\n",
       "      <td>5.387542</td>\n",
       "      <td>12.988765</td>\n",
       "      <td>2.752951</td>\n",
       "      <td>10.808698</td>\n",
       "      <td>6.783006</td>\n",
       "      <td>...</td>\n",
       "      <td>8.173266</td>\n",
       "      <td>2.920829</td>\n",
       "      <td>6.195751</td>\n",
       "      <td>3.607377</td>\n",
       "      <td>7.433029</td>\n",
       "      <td>5.071759</td>\n",
       "      <td>4.378420</td>\n",
       "      <td>5.218668</td>\n",
       "      <td>12.032507</td>\n",
       "      <td>4.576603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067438</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.109171</td>\n",
       "      <td>5.186781</td>\n",
       "      <td>3.728779</td>\n",
       "      <td>12.374821</td>\n",
       "      <td>2.752656</td>\n",
       "      <td>10.174804</td>\n",
       "      <td>6.352175</td>\n",
       "      <td>...</td>\n",
       "      <td>8.651848</td>\n",
       "      <td>5.168417</td>\n",
       "      <td>5.846072</td>\n",
       "      <td>3.281763</td>\n",
       "      <td>6.708472</td>\n",
       "      <td>6.498593</td>\n",
       "      <td>3.838492</td>\n",
       "      <td>6.246277</td>\n",
       "      <td>12.525255</td>\n",
       "      <td>5.089923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067439</th>\n",
       "      <td>0.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.194998</td>\n",
       "      <td>5.993580</td>\n",
       "      <td>2.627080</td>\n",
       "      <td>11.902411</td>\n",
       "      <td>3.744308</td>\n",
       "      <td>11.807810</td>\n",
       "      <td>5.707255</td>\n",
       "      <td>...</td>\n",
       "      <td>7.784618</td>\n",
       "      <td>2.620130</td>\n",
       "      <td>5.676708</td>\n",
       "      <td>4.002145</td>\n",
       "      <td>7.640616</td>\n",
       "      <td>2.766557</td>\n",
       "      <td>2.780148</td>\n",
       "      <td>2.793083</td>\n",
       "      <td>11.593306</td>\n",
       "      <td>2.836373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067440</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.563399</td>\n",
       "      <td>5.092242</td>\n",
       "      <td>3.959856</td>\n",
       "      <td>11.898668</td>\n",
       "      <td>3.207581</td>\n",
       "      <td>10.781315</td>\n",
       "      <td>6.238998</td>\n",
       "      <td>...</td>\n",
       "      <td>8.309621</td>\n",
       "      <td>3.658243</td>\n",
       "      <td>6.139664</td>\n",
       "      <td>4.686072</td>\n",
       "      <td>7.115034</td>\n",
       "      <td>5.351994</td>\n",
       "      <td>2.796012</td>\n",
       "      <td>5.975166</td>\n",
       "      <td>12.481041</td>\n",
       "      <td>3.461828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067441</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.781489</td>\n",
       "      <td>4.207420</td>\n",
       "      <td>5.515544</td>\n",
       "      <td>12.284374</td>\n",
       "      <td>3.985785</td>\n",
       "      <td>10.875250</td>\n",
       "      <td>6.499204</td>\n",
       "      <td>...</td>\n",
       "      <td>9.144845</td>\n",
       "      <td>4.721522</td>\n",
       "      <td>6.197854</td>\n",
       "      <td>4.055265</td>\n",
       "      <td>7.025485</td>\n",
       "      <td>6.635454</td>\n",
       "      <td>3.279048</td>\n",
       "      <td>6.454967</td>\n",
       "      <td>13.381455</td>\n",
       "      <td>4.433457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067442</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.850274</td>\n",
       "      <td>4.827898</td>\n",
       "      <td>2.767853</td>\n",
       "      <td>11.571885</td>\n",
       "      <td>2.860440</td>\n",
       "      <td>10.859670</td>\n",
       "      <td>6.217347</td>\n",
       "      <td>...</td>\n",
       "      <td>8.236813</td>\n",
       "      <td>3.106348</td>\n",
       "      <td>5.792785</td>\n",
       "      <td>4.603995</td>\n",
       "      <td>7.067878</td>\n",
       "      <td>4.963118</td>\n",
       "      <td>4.103069</td>\n",
       "      <td>4.314279</td>\n",
       "      <td>12.934966</td>\n",
       "      <td>5.089310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067443</th>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.471855</td>\n",
       "      <td>4.428555</td>\n",
       "      <td>5.630373</td>\n",
       "      <td>13.675087</td>\n",
       "      <td>4.050741</td>\n",
       "      <td>10.786637</td>\n",
       "      <td>6.176792</td>\n",
       "      <td>...</td>\n",
       "      <td>8.415085</td>\n",
       "      <td>4.536071</td>\n",
       "      <td>6.250094</td>\n",
       "      <td>4.105694</td>\n",
       "      <td>7.265487</td>\n",
       "      <td>6.386533</td>\n",
       "      <td>2.784441</td>\n",
       "      <td>5.490700</td>\n",
       "      <td>12.242868</td>\n",
       "      <td>5.473535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067444</th>\n",
       "      <td>0.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.437905</td>\n",
       "      <td>5.144545</td>\n",
       "      <td>3.992799</td>\n",
       "      <td>13.188519</td>\n",
       "      <td>3.471841</td>\n",
       "      <td>10.787437</td>\n",
       "      <td>6.455756</td>\n",
       "      <td>...</td>\n",
       "      <td>8.694051</td>\n",
       "      <td>3.821267</td>\n",
       "      <td>6.465984</td>\n",
       "      <td>5.017246</td>\n",
       "      <td>7.009344</td>\n",
       "      <td>6.044277</td>\n",
       "      <td>2.844123</td>\n",
       "      <td>6.438496</td>\n",
       "      <td>12.662624</td>\n",
       "      <td>5.340743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067445</th>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.139027</td>\n",
       "      <td>4.265222</td>\n",
       "      <td>3.704854</td>\n",
       "      <td>12.212286</td>\n",
       "      <td>2.683592</td>\n",
       "      <td>10.730342</td>\n",
       "      <td>6.310343</td>\n",
       "      <td>...</td>\n",
       "      <td>8.719358</td>\n",
       "      <td>3.171274</td>\n",
       "      <td>5.962624</td>\n",
       "      <td>4.347804</td>\n",
       "      <td>6.648044</td>\n",
       "      <td>6.197403</td>\n",
       "      <td>3.001317</td>\n",
       "      <td>5.938834</td>\n",
       "      <td>13.345444</td>\n",
       "      <td>4.358263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067446</th>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.879640</td>\n",
       "      <td>4.939137</td>\n",
       "      <td>4.966670</td>\n",
       "      <td>11.598589</td>\n",
       "      <td>3.042314</td>\n",
       "      <td>10.701382</td>\n",
       "      <td>6.307161</td>\n",
       "      <td>...</td>\n",
       "      <td>8.591033</td>\n",
       "      <td>3.740378</td>\n",
       "      <td>5.936614</td>\n",
       "      <td>3.777043</td>\n",
       "      <td>6.953502</td>\n",
       "      <td>6.234549</td>\n",
       "      <td>3.254741</td>\n",
       "      <td>6.089719</td>\n",
       "      <td>12.937860</td>\n",
       "      <td>4.567399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067447</th>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.854952</td>\n",
       "      <td>3.701736</td>\n",
       "      <td>5.430852</td>\n",
       "      <td>11.697882</td>\n",
       "      <td>3.493999</td>\n",
       "      <td>11.321520</td>\n",
       "      <td>6.474544</td>\n",
       "      <td>...</td>\n",
       "      <td>8.429256</td>\n",
       "      <td>4.774469</td>\n",
       "      <td>6.036672</td>\n",
       "      <td>4.654058</td>\n",
       "      <td>7.628637</td>\n",
       "      <td>6.900276</td>\n",
       "      <td>3.443353</td>\n",
       "      <td>6.119936</td>\n",
       "      <td>12.877521</td>\n",
       "      <td>5.118452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067448</th>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.728254</td>\n",
       "      <td>3.883522</td>\n",
       "      <td>6.139148</td>\n",
       "      <td>11.988255</td>\n",
       "      <td>2.737365</td>\n",
       "      <td>11.107452</td>\n",
       "      <td>6.219859</td>\n",
       "      <td>...</td>\n",
       "      <td>8.530454</td>\n",
       "      <td>5.083735</td>\n",
       "      <td>6.270426</td>\n",
       "      <td>4.881639</td>\n",
       "      <td>7.537494</td>\n",
       "      <td>6.748971</td>\n",
       "      <td>3.214889</td>\n",
       "      <td>6.435175</td>\n",
       "      <td>12.797197</td>\n",
       "      <td>5.146932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067449</th>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.090071</td>\n",
       "      <td>4.024209</td>\n",
       "      <td>3.948836</td>\n",
       "      <td>12.683393</td>\n",
       "      <td>4.080232</td>\n",
       "      <td>10.625716</td>\n",
       "      <td>6.199072</td>\n",
       "      <td>...</td>\n",
       "      <td>8.298539</td>\n",
       "      <td>4.249813</td>\n",
       "      <td>5.744754</td>\n",
       "      <td>3.822602</td>\n",
       "      <td>6.868358</td>\n",
       "      <td>6.278687</td>\n",
       "      <td>2.938344</td>\n",
       "      <td>5.882005</td>\n",
       "      <td>12.584850</td>\n",
       "      <td>4.554525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067450</th>\n",
       "      <td>0.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.510730</td>\n",
       "      <td>4.110143</td>\n",
       "      <td>5.609548</td>\n",
       "      <td>12.474763</td>\n",
       "      <td>3.875072</td>\n",
       "      <td>10.728470</td>\n",
       "      <td>6.006374</td>\n",
       "      <td>...</td>\n",
       "      <td>8.927208</td>\n",
       "      <td>4.798983</td>\n",
       "      <td>6.251776</td>\n",
       "      <td>4.361403</td>\n",
       "      <td>7.024695</td>\n",
       "      <td>6.725098</td>\n",
       "      <td>2.809439</td>\n",
       "      <td>6.615240</td>\n",
       "      <td>13.253766</td>\n",
       "      <td>5.249892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067451</th>\n",
       "      <td>0.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.516667</td>\n",
       "      <td>3.200867</td>\n",
       "      <td>6.374747</td>\n",
       "      <td>12.852168</td>\n",
       "      <td>3.207471</td>\n",
       "      <td>11.085492</td>\n",
       "      <td>6.119696</td>\n",
       "      <td>...</td>\n",
       "      <td>9.109817</td>\n",
       "      <td>5.744519</td>\n",
       "      <td>6.745788</td>\n",
       "      <td>5.718458</td>\n",
       "      <td>7.326384</td>\n",
       "      <td>7.014871</td>\n",
       "      <td>2.832272</td>\n",
       "      <td>6.468919</td>\n",
       "      <td>12.833095</td>\n",
       "      <td>5.960049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067452</th>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.717634</td>\n",
       "      <td>3.770461</td>\n",
       "      <td>5.868522</td>\n",
       "      <td>11.945599</td>\n",
       "      <td>2.937372</td>\n",
       "      <td>11.067977</td>\n",
       "      <td>6.099312</td>\n",
       "      <td>...</td>\n",
       "      <td>8.920365</td>\n",
       "      <td>5.480914</td>\n",
       "      <td>6.728947</td>\n",
       "      <td>5.588893</td>\n",
       "      <td>7.449795</td>\n",
       "      <td>6.676802</td>\n",
       "      <td>2.801748</td>\n",
       "      <td>6.733632</td>\n",
       "      <td>13.070649</td>\n",
       "      <td>6.031067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067453</th>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.436784</td>\n",
       "      <td>4.451320</td>\n",
       "      <td>6.042601</td>\n",
       "      <td>12.211297</td>\n",
       "      <td>3.411104</td>\n",
       "      <td>10.777494</td>\n",
       "      <td>6.151203</td>\n",
       "      <td>...</td>\n",
       "      <td>9.027000</td>\n",
       "      <td>4.625002</td>\n",
       "      <td>6.427583</td>\n",
       "      <td>4.688361</td>\n",
       "      <td>7.119568</td>\n",
       "      <td>6.647157</td>\n",
       "      <td>2.735566</td>\n",
       "      <td>6.347763</td>\n",
       "      <td>13.338604</td>\n",
       "      <td>5.386587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067454</th>\n",
       "      <td>0.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.354268</td>\n",
       "      <td>4.461054</td>\n",
       "      <td>5.190834</td>\n",
       "      <td>11.890785</td>\n",
       "      <td>3.899433</td>\n",
       "      <td>10.744350</td>\n",
       "      <td>6.188550</td>\n",
       "      <td>...</td>\n",
       "      <td>8.637108</td>\n",
       "      <td>4.656932</td>\n",
       "      <td>6.397263</td>\n",
       "      <td>5.305038</td>\n",
       "      <td>6.965261</td>\n",
       "      <td>6.405932</td>\n",
       "      <td>2.748930</td>\n",
       "      <td>6.040958</td>\n",
       "      <td>12.802153</td>\n",
       "      <td>5.418139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067455</th>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.692356</td>\n",
       "      <td>4.306510</td>\n",
       "      <td>5.677143</td>\n",
       "      <td>12.882605</td>\n",
       "      <td>4.059085</td>\n",
       "      <td>10.850011</td>\n",
       "      <td>6.112434</td>\n",
       "      <td>...</td>\n",
       "      <td>8.538701</td>\n",
       "      <td>4.119037</td>\n",
       "      <td>6.159858</td>\n",
       "      <td>4.060721</td>\n",
       "      <td>7.422848</td>\n",
       "      <td>6.357309</td>\n",
       "      <td>2.876586</td>\n",
       "      <td>6.040609</td>\n",
       "      <td>12.110378</td>\n",
       "      <td>4.437177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067456</th>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.509643</td>\n",
       "      <td>3.974989</td>\n",
       "      <td>5.098562</td>\n",
       "      <td>12.581439</td>\n",
       "      <td>2.905740</td>\n",
       "      <td>10.599248</td>\n",
       "      <td>6.202716</td>\n",
       "      <td>...</td>\n",
       "      <td>8.612995</td>\n",
       "      <td>4.759997</td>\n",
       "      <td>6.397822</td>\n",
       "      <td>4.928984</td>\n",
       "      <td>7.265245</td>\n",
       "      <td>6.456726</td>\n",
       "      <td>2.828128</td>\n",
       "      <td>6.360076</td>\n",
       "      <td>12.578838</td>\n",
       "      <td>5.433254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067457</th>\n",
       "      <td>0.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.310910</td>\n",
       "      <td>3.481818</td>\n",
       "      <td>5.875097</td>\n",
       "      <td>11.951515</td>\n",
       "      <td>2.999079</td>\n",
       "      <td>11.128664</td>\n",
       "      <td>6.262343</td>\n",
       "      <td>...</td>\n",
       "      <td>8.452632</td>\n",
       "      <td>4.796967</td>\n",
       "      <td>6.129514</td>\n",
       "      <td>5.339343</td>\n",
       "      <td>7.627941</td>\n",
       "      <td>6.496564</td>\n",
       "      <td>2.962675</td>\n",
       "      <td>6.197123</td>\n",
       "      <td>12.872514</td>\n",
       "      <td>5.885852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067458</th>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.487669</td>\n",
       "      <td>4.010231</td>\n",
       "      <td>5.961306</td>\n",
       "      <td>11.085049</td>\n",
       "      <td>3.518736</td>\n",
       "      <td>10.769642</td>\n",
       "      <td>6.175364</td>\n",
       "      <td>...</td>\n",
       "      <td>8.783906</td>\n",
       "      <td>4.111435</td>\n",
       "      <td>6.000232</td>\n",
       "      <td>4.155737</td>\n",
       "      <td>6.865086</td>\n",
       "      <td>6.893305</td>\n",
       "      <td>2.874528</td>\n",
       "      <td>5.809800</td>\n",
       "      <td>12.954499</td>\n",
       "      <td>4.147930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067459</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.933303</td>\n",
       "      <td>4.262363</td>\n",
       "      <td>5.425658</td>\n",
       "      <td>12.906422</td>\n",
       "      <td>3.275482</td>\n",
       "      <td>11.048617</td>\n",
       "      <td>6.665316</td>\n",
       "      <td>...</td>\n",
       "      <td>8.202754</td>\n",
       "      <td>4.224158</td>\n",
       "      <td>5.600002</td>\n",
       "      <td>4.370410</td>\n",
       "      <td>7.675758</td>\n",
       "      <td>6.206098</td>\n",
       "      <td>2.942508</td>\n",
       "      <td>5.440079</td>\n",
       "      <td>12.288324</td>\n",
       "      <td>4.563690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067460</th>\n",
       "      <td>0.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.728533</td>\n",
       "      <td>4.172008</td>\n",
       "      <td>5.313376</td>\n",
       "      <td>12.599278</td>\n",
       "      <td>2.708914</td>\n",
       "      <td>11.095754</td>\n",
       "      <td>6.400799</td>\n",
       "      <td>...</td>\n",
       "      <td>8.285680</td>\n",
       "      <td>3.230213</td>\n",
       "      <td>5.604343</td>\n",
       "      <td>4.359566</td>\n",
       "      <td>7.536837</td>\n",
       "      <td>6.251587</td>\n",
       "      <td>3.252552</td>\n",
       "      <td>5.502089</td>\n",
       "      <td>12.264850</td>\n",
       "      <td>4.904048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067461</th>\n",
       "      <td>0.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.522120</td>\n",
       "      <td>3.728118</td>\n",
       "      <td>6.113496</td>\n",
       "      <td>12.247601</td>\n",
       "      <td>2.969698</td>\n",
       "      <td>10.948413</td>\n",
       "      <td>6.259489</td>\n",
       "      <td>...</td>\n",
       "      <td>8.843767</td>\n",
       "      <td>5.073464</td>\n",
       "      <td>6.331352</td>\n",
       "      <td>5.012224</td>\n",
       "      <td>7.171527</td>\n",
       "      <td>6.796412</td>\n",
       "      <td>2.821534</td>\n",
       "      <td>6.425600</td>\n",
       "      <td>13.421458</td>\n",
       "      <td>5.616683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2067462</th>\n",
       "      <td>0.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.859007</td>\n",
       "      <td>3.784718</td>\n",
       "      <td>5.387186</td>\n",
       "      <td>12.555675</td>\n",
       "      <td>3.287521</td>\n",
       "      <td>10.933269</td>\n",
       "      <td>6.209529</td>\n",
       "      <td>...</td>\n",
       "      <td>8.606899</td>\n",
       "      <td>4.663775</td>\n",
       "      <td>6.160120</td>\n",
       "      <td>4.493533</td>\n",
       "      <td>7.292884</td>\n",
       "      <td>6.615562</td>\n",
       "      <td>3.412952</td>\n",
       "      <td>6.460133</td>\n",
       "      <td>12.969342</td>\n",
       "      <td>5.282989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60 rows Ã— 14032 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Anxiety disorder   Age  Gender      A1BG      A1CF       A2M  \\\n",
       "GSM2067403               1.0  25.0     0.0  6.714729  3.938725  4.105190   \n",
       "GSM2067404               1.0  23.0     1.0  6.119238  3.857275  5.898689   \n",
       "GSM2067405               1.0  18.0     1.0  6.588651  4.320075  5.163733   \n",
       "GSM2067406               1.0  26.0     0.0  6.389377  4.589062  5.968357   \n",
       "GSM2067407               1.0  27.0     1.0  6.489448  4.668073  4.846476   \n",
       "GSM2067408               1.0  19.0     0.0  6.229496  4.644432  5.877680   \n",
       "GSM2067409               1.0  22.0     0.0  5.642367  6.507755  5.523924   \n",
       "GSM2067410               1.0  27.0     1.0  5.565992  5.445320  6.222440   \n",
       "GSM2067411               1.0  18.0     0.0  6.170604  3.228480  6.007900   \n",
       "GSM2067412               1.0  25.0     0.0  6.437280  4.336737  6.735513   \n",
       "GSM2067413               1.0  16.0     0.0  5.624097  4.086302  5.618996   \n",
       "GSM2067414               1.0  35.0     1.0  6.508125  4.479011  5.697313   \n",
       "GSM2067415               1.0  16.0     0.0  6.395001  3.703604  5.590061   \n",
       "GSM2067416               1.0  16.0     0.0  6.398784  4.504799  5.234035   \n",
       "GSM2067417               1.0  32.0     0.0  6.955859  3.523432  6.458198   \n",
       "GSM2067418               1.0  18.0     0.0  6.926815  2.940408  6.273612   \n",
       "GSM2067419               1.0  15.0     0.0  6.700031  3.603868  6.608646   \n",
       "GSM2067420               1.0  43.0     0.0  6.420770  4.308397  4.419219   \n",
       "GSM2067421               1.0  36.0     0.0  6.865487  3.445940  5.425325   \n",
       "GSM2067422               1.0  17.0     1.0  6.839920  3.690498  6.696085   \n",
       "GSM2067423               1.0  45.0     1.0  6.770218  4.635688  5.216162   \n",
       "GSM2067424               1.0  40.0     0.0  6.359819  4.801514  4.932057   \n",
       "GSM2067425               1.0  35.0     0.0  6.663395  3.894671  5.955735   \n",
       "GSM2067426               1.0  28.0     1.0  6.663239  4.511771  5.806159   \n",
       "GSM2067427               1.0  27.0     0.0  6.542545  4.366632  5.442149   \n",
       "GSM2067428               1.0  31.0     0.0  6.877935  3.891802  4.252012   \n",
       "GSM2067429               1.0  23.0     0.0  6.517502  4.347663  5.699042   \n",
       "GSM2067430               1.0  35.0     1.0  7.100269  4.299640  4.012525   \n",
       "GSM2067431               1.0  60.0     0.0  6.467290  4.048092  5.393207   \n",
       "GSM2067432               1.0  59.0     1.0  6.671121  4.381368  6.989244   \n",
       "GSM2067433               0.0  24.0     0.0  6.650069  3.928323  5.469918   \n",
       "GSM2067434               0.0  23.0     1.0  6.990864  4.462671  6.327537   \n",
       "GSM2067435               0.0  18.0     1.0  6.717298  3.816540  6.621448   \n",
       "GSM2067436               0.0  27.0     0.0  6.055410  5.362066  3.588085   \n",
       "GSM2067437               0.0  27.0     1.0  6.788504  4.898884  5.387542   \n",
       "GSM2067438               0.0  20.0     0.0  5.109171  5.186781  3.728779   \n",
       "GSM2067439               0.0  21.0     0.0  7.194998  5.993580  2.627080   \n",
       "GSM2067440               0.0  27.0     1.0  5.563399  5.092242  3.959856   \n",
       "GSM2067441               0.0  20.0     0.0  6.781489  4.207420  5.515544   \n",
       "GSM2067442               0.0  24.0     0.0  5.850274  4.827898  2.767853   \n",
       "GSM2067443               0.0  18.0     0.0  6.471855  4.428555  5.630373   \n",
       "GSM2067444               0.0  35.0     1.0  6.437905  5.144545  3.992799   \n",
       "GSM2067445               0.0  17.0     0.0  6.139027  4.265222  3.704854   \n",
       "GSM2067446               0.0  18.0     0.0  5.879640  4.939137  4.966670   \n",
       "GSM2067447               0.0  32.0     0.0  6.854952  3.701736  5.430852   \n",
       "GSM2067448               0.0  18.0     0.0  6.728254  3.883522  6.139148   \n",
       "GSM2067449               0.0  18.0     0.0  6.090071  4.024209  3.948836   \n",
       "GSM2067450               0.0  44.0     0.0  6.510730  4.110143  5.609548   \n",
       "GSM2067451               0.0  37.0     0.0  6.516667  3.200867  6.374747   \n",
       "GSM2067452               0.0  17.0     1.0  6.717634  3.770461  5.868522   \n",
       "GSM2067453               0.0  43.0     1.0  6.436784  4.451320  6.042601   \n",
       "GSM2067454               0.0  40.0     0.0  6.354268  4.461054  5.190834   \n",
       "GSM2067455               0.0  32.0     0.0  6.692356  4.306510  5.677143   \n",
       "GSM2067456               0.0  28.0     1.0  6.509643  3.974989  5.098562   \n",
       "GSM2067457               0.0  27.0     0.0  7.310910  3.481818  5.875097   \n",
       "GSM2067458               0.0  30.0     0.0  6.487669  4.010231  5.961306   \n",
       "GSM2067459               0.0  24.0     0.0  6.933303  4.262363  5.425658   \n",
       "GSM2067460               0.0  35.0     1.0  6.728533  4.172008  5.313376   \n",
       "GSM2067461               0.0  56.0     0.0  6.522120  3.728118  6.113496   \n",
       "GSM2067462               0.0  56.0     1.0  6.859007  3.784718  5.387186   \n",
       "\n",
       "               A4GALT     A4GNT       AAAS      AACS  ...    ZSWIM6    ZSWIM7  \\\n",
       "GSM2067403  13.030699  2.704922  11.219065  6.193231  ...  8.472954  4.146875   \n",
       "GSM2067404  12.865675  2.733084  10.718920  5.529508  ...  8.837225  5.017093   \n",
       "GSM2067405  13.568920  3.909408  11.084585  6.278926  ...  8.916005  5.166640   \n",
       "GSM2067406  12.454192  2.893169  10.745497  5.896101  ...  8.870725  4.644601   \n",
       "GSM2067407  13.582437  2.692215  10.799071  6.265001  ...  8.773541  3.564052   \n",
       "GSM2067408  12.652796  3.176106   9.883424  6.285170  ...  8.755016  5.112708   \n",
       "GSM2067409  12.800940  3.038265  10.526562  6.054632  ...  8.459496  5.665781   \n",
       "GSM2067410  11.776719  4.287126  10.491643  6.232655  ...  8.678071  4.061905   \n",
       "GSM2067411  12.721724  2.768476  11.095754  6.630271  ...  8.630165  4.591509   \n",
       "GSM2067412  12.280159  2.787102  10.937666  6.262236  ...  8.848257  4.779469   \n",
       "GSM2067413  13.730714  3.903112  10.699893  6.376582  ...  8.572907  5.004915   \n",
       "GSM2067414  13.405120  3.854834  10.876848  6.166716  ...  8.586086  3.447876   \n",
       "GSM2067415  12.015218  3.254557  10.477460  6.270210  ...  8.996468  3.963160   \n",
       "GSM2067416  12.138115  3.756422  10.797807  6.289853  ...  8.771525  4.398438   \n",
       "GSM2067417  12.381153  3.165949  11.393041  6.272181  ...  8.896212  5.605754   \n",
       "GSM2067418  12.155148  3.010371  11.172894  6.242284  ...  9.155986  5.731468   \n",
       "GSM2067419  13.373009  3.368898  11.111868  6.140389  ...  8.790861  4.912009   \n",
       "GSM2067420  12.657779  4.182192  10.265317  5.575451  ...  8.938309  4.391050   \n",
       "GSM2067421  12.538816  3.530409  10.940994  6.028025  ...  8.903367  5.238989   \n",
       "GSM2067422  12.435116  3.284214  11.218464  6.190138  ...  8.994142  5.549764   \n",
       "GSM2067423  12.126523  4.427112  10.802723  6.136266  ...  8.627634  4.336113   \n",
       "GSM2067424  12.145253  3.818097  10.907265  6.121207  ...  8.987240  4.889013   \n",
       "GSM2067425  13.277133  2.958530  10.570036  5.945974  ...  8.630093  4.538773   \n",
       "GSM2067426  12.673718  3.400782  10.956592  6.158459  ...  8.758475  4.394046   \n",
       "GSM2067427  12.275525  2.735066  11.355851  6.237150  ...  9.033045  3.229970   \n",
       "GSM2067428  11.616556  3.679624  10.997719  5.853625  ...  9.273019  5.005182   \n",
       "GSM2067429  12.834258  3.219715  10.688426  5.963651  ...  8.611595  4.199718   \n",
       "GSM2067430  13.097274  3.685450  10.906361  6.371967  ...  8.248276  3.136578   \n",
       "GSM2067431  13.331580  3.237607   9.928258  5.895905  ...  9.171186  5.097364   \n",
       "GSM2067432  12.360163  3.558934  11.041781  6.190573  ...  8.743233  4.372523   \n",
       "GSM2067433  13.393532  2.809327  10.821529  6.230222  ...  8.686610  5.212686   \n",
       "GSM2067434  13.056033  2.787889  10.704703  5.867151  ...  8.755357  4.877190   \n",
       "GSM2067435  13.331580  3.488667  10.716845  5.750632  ...  8.492819  5.080455   \n",
       "GSM2067436  12.241908  3.464015  10.516702  6.380311  ...  8.438438  4.235894   \n",
       "GSM2067437  12.988765  2.752951  10.808698  6.783006  ...  8.173266  2.920829   \n",
       "GSM2067438  12.374821  2.752656  10.174804  6.352175  ...  8.651848  5.168417   \n",
       "GSM2067439  11.902411  3.744308  11.807810  5.707255  ...  7.784618  2.620130   \n",
       "GSM2067440  11.898668  3.207581  10.781315  6.238998  ...  8.309621  3.658243   \n",
       "GSM2067441  12.284374  3.985785  10.875250  6.499204  ...  9.144845  4.721522   \n",
       "GSM2067442  11.571885  2.860440  10.859670  6.217347  ...  8.236813  3.106348   \n",
       "GSM2067443  13.675087  4.050741  10.786637  6.176792  ...  8.415085  4.536071   \n",
       "GSM2067444  13.188519  3.471841  10.787437  6.455756  ...  8.694051  3.821267   \n",
       "GSM2067445  12.212286  2.683592  10.730342  6.310343  ...  8.719358  3.171274   \n",
       "GSM2067446  11.598589  3.042314  10.701382  6.307161  ...  8.591033  3.740378   \n",
       "GSM2067447  11.697882  3.493999  11.321520  6.474544  ...  8.429256  4.774469   \n",
       "GSM2067448  11.988255  2.737365  11.107452  6.219859  ...  8.530454  5.083735   \n",
       "GSM2067449  12.683393  4.080232  10.625716  6.199072  ...  8.298539  4.249813   \n",
       "GSM2067450  12.474763  3.875072  10.728470  6.006374  ...  8.927208  4.798983   \n",
       "GSM2067451  12.852168  3.207471  11.085492  6.119696  ...  9.109817  5.744519   \n",
       "GSM2067452  11.945599  2.937372  11.067977  6.099312  ...  8.920365  5.480914   \n",
       "GSM2067453  12.211297  3.411104  10.777494  6.151203  ...  9.027000  4.625002   \n",
       "GSM2067454  11.890785  3.899433  10.744350  6.188550  ...  8.637108  4.656932   \n",
       "GSM2067455  12.882605  4.059085  10.850011  6.112434  ...  8.538701  4.119037   \n",
       "GSM2067456  12.581439  2.905740  10.599248  6.202716  ...  8.612995  4.759997   \n",
       "GSM2067457  11.951515  2.999079  11.128664  6.262343  ...  8.452632  4.796967   \n",
       "GSM2067458  11.085049  3.518736  10.769642  6.175364  ...  8.783906  4.111435   \n",
       "GSM2067459  12.906422  3.275482  11.048617  6.665316  ...  8.202754  4.224158   \n",
       "GSM2067460  12.599278  2.708914  11.095754  6.400799  ...  8.285680  3.230213   \n",
       "GSM2067461  12.247601  2.969698  10.948413  6.259489  ...  8.843767  5.073464   \n",
       "GSM2067462  12.555675  3.287521  10.933269  6.209529  ...  8.606899  4.663775   \n",
       "\n",
       "                ZW10    ZWILCH      ZXDA      ZXDC    ZYG11A    ZYG11B  \\\n",
       "GSM2067403  5.884128  3.879921  7.125728  6.206871  3.020767  5.951725   \n",
       "GSM2067404  6.116274  4.135404  7.121595  6.407975  3.026273  6.141165   \n",
       "GSM2067405  5.942174  2.728597  7.344893  6.309663  3.113522  5.639593   \n",
       "GSM2067406  5.981628  4.986273  7.135086  6.479846  2.773175  6.538147   \n",
       "GSM2067407  6.142061  4.558930  7.214107  6.401652  2.981554  5.849266   \n",
       "GSM2067408  6.385875  4.676870  6.742880  6.667216  3.513956  7.026554   \n",
       "GSM2067409  6.182136  4.099351  6.705152  6.080510  3.391591  5.296118   \n",
       "GSM2067410  5.991161  4.052005  7.071524  5.717164  3.025322  5.633491   \n",
       "GSM2067411  6.099095  3.889528  6.957309  6.664669  3.290877  6.056791   \n",
       "GSM2067412  6.101059  4.707293  7.179975  5.979232  3.548486  6.068389   \n",
       "GSM2067413  5.937257  5.041395  6.972745  6.073244  2.817874  6.416759   \n",
       "GSM2067414  6.369736  3.533126  7.343472  6.380209  2.848518  5.788387   \n",
       "GSM2067415  5.912210  4.288434  6.621121  6.607335  3.140500  6.342705   \n",
       "GSM2067416  6.232646  3.979691  6.619864  6.575380  3.098263  5.728942   \n",
       "GSM2067417  6.731539  5.507618  7.359767  7.044290  3.271596  6.527856   \n",
       "GSM2067418  6.437933  5.371328  7.483693  7.304184  2.870487  7.111848   \n",
       "GSM2067419  6.122825  4.528871  7.217832  6.580385  2.890108  6.398405   \n",
       "GSM2067420  5.829993  3.814391  6.844831  6.732603  2.956524  7.408227   \n",
       "GSM2067421  6.509625  5.306362  7.448940  7.041186  2.953032  6.538046   \n",
       "GSM2067422  6.738438  5.249587  7.580654  6.882779  2.926269  6.844660   \n",
       "GSM2067423  5.497256  4.016527  7.139729  6.044130  2.774483  5.807989   \n",
       "GSM2067424  6.417872  4.120697  6.983198  6.486918  2.733062  6.645017   \n",
       "GSM2067425  6.209540  4.679893  7.238637  6.523522  3.026820  6.441123   \n",
       "GSM2067426  6.324679  4.530297  7.257558  6.820784  2.899090  6.546835   \n",
       "GSM2067427  5.884744  4.018757  7.167019  6.326521  3.125467  6.801677   \n",
       "GSM2067428  5.920124  4.349660  6.814784  7.202935  3.026861  6.738949   \n",
       "GSM2067429  5.700428  4.972825  7.023918  6.558665  2.869833  6.466369   \n",
       "GSM2067430  5.498039  4.274403  7.583811  5.958409  3.068320  5.905375   \n",
       "GSM2067431  6.366438  6.166522  6.801190  7.086836  2.805352  7.007291   \n",
       "GSM2067432  5.765166  4.768969  7.329707  6.971662  2.899214  6.184332   \n",
       "GSM2067433  6.449946  5.164700  6.903481  6.541790  3.213683  6.583720   \n",
       "GSM2067434  6.319072  4.974823  7.240470  5.733144  3.366409  6.213997   \n",
       "GSM2067435  6.644737  4.943316  7.308149  6.303049  2.803965  6.304306   \n",
       "GSM2067436  5.786593  4.708460  6.783018  5.907134  2.986399  6.238057   \n",
       "GSM2067437  6.195751  3.607377  7.433029  5.071759  4.378420  5.218668   \n",
       "GSM2067438  5.846072  3.281763  6.708472  6.498593  3.838492  6.246277   \n",
       "GSM2067439  5.676708  4.002145  7.640616  2.766557  2.780148  2.793083   \n",
       "GSM2067440  6.139664  4.686072  7.115034  5.351994  2.796012  5.975166   \n",
       "GSM2067441  6.197854  4.055265  7.025485  6.635454  3.279048  6.454967   \n",
       "GSM2067442  5.792785  4.603995  7.067878  4.963118  4.103069  4.314279   \n",
       "GSM2067443  6.250094  4.105694  7.265487  6.386533  2.784441  5.490700   \n",
       "GSM2067444  6.465984  5.017246  7.009344  6.044277  2.844123  6.438496   \n",
       "GSM2067445  5.962624  4.347804  6.648044  6.197403  3.001317  5.938834   \n",
       "GSM2067446  5.936614  3.777043  6.953502  6.234549  3.254741  6.089719   \n",
       "GSM2067447  6.036672  4.654058  7.628637  6.900276  3.443353  6.119936   \n",
       "GSM2067448  6.270426  4.881639  7.537494  6.748971  3.214889  6.435175   \n",
       "GSM2067449  5.744754  3.822602  6.868358  6.278687  2.938344  5.882005   \n",
       "GSM2067450  6.251776  4.361403  7.024695  6.725098  2.809439  6.615240   \n",
       "GSM2067451  6.745788  5.718458  7.326384  7.014871  2.832272  6.468919   \n",
       "GSM2067452  6.728947  5.588893  7.449795  6.676802  2.801748  6.733632   \n",
       "GSM2067453  6.427583  4.688361  7.119568  6.647157  2.735566  6.347763   \n",
       "GSM2067454  6.397263  5.305038  6.965261  6.405932  2.748930  6.040958   \n",
       "GSM2067455  6.159858  4.060721  7.422848  6.357309  2.876586  6.040609   \n",
       "GSM2067456  6.397822  4.928984  7.265245  6.456726  2.828128  6.360076   \n",
       "GSM2067457  6.129514  5.339343  7.627941  6.496564  2.962675  6.197123   \n",
       "GSM2067458  6.000232  4.155737  6.865086  6.893305  2.874528  5.809800   \n",
       "GSM2067459  5.600002  4.370410  7.675758  6.206098  2.942508  5.440079   \n",
       "GSM2067460  5.604343  4.359566  7.536837  6.251587  3.252552  5.502089   \n",
       "GSM2067461  6.331352  5.012224  7.171527  6.796412  2.821534  6.425600   \n",
       "GSM2067462  6.160120  4.493533  7.292884  6.615562  3.412952  6.460133   \n",
       "\n",
       "                  ZYX      ZZZ3  \n",
       "GSM2067403  12.271689  3.938076  \n",
       "GSM2067404  12.290836  4.905338  \n",
       "GSM2067405  12.674829  4.189789  \n",
       "GSM2067406  13.083159  5.280000  \n",
       "GSM2067407  12.931067  4.669816  \n",
       "GSM2067408  12.483398  5.285511  \n",
       "GSM2067409  12.510457  5.079488  \n",
       "GSM2067410  12.603576  4.814769  \n",
       "GSM2067411  13.032246  4.740775  \n",
       "GSM2067412  13.294943  4.656932  \n",
       "GSM2067413  12.373903  5.321508  \n",
       "GSM2067414  12.820009  4.926111  \n",
       "GSM2067415  12.940755  4.109571  \n",
       "GSM2067416  13.071623  4.070495  \n",
       "GSM2067417  13.238666  5.625271  \n",
       "GSM2067418  13.220077  5.601524  \n",
       "GSM2067419  12.554396  5.229794  \n",
       "GSM2067420  12.217965  4.816319  \n",
       "GSM2067421  12.637005  5.844087  \n",
       "GSM2067422  13.024989  5.728438  \n",
       "GSM2067423  12.337760  4.647233  \n",
       "GSM2067424  12.957374  4.746378  \n",
       "GSM2067425  12.399323  5.098562  \n",
       "GSM2067426  12.772914  5.087595  \n",
       "GSM2067427  13.077705  4.810358  \n",
       "GSM2067428  12.957374  5.203513  \n",
       "GSM2067429  11.803370  5.039581  \n",
       "GSM2067430  11.859551  5.343675  \n",
       "GSM2067431  13.104760  5.478596  \n",
       "GSM2067432  13.217264  4.937017  \n",
       "GSM2067433  12.431108  5.720720  \n",
       "GSM2067434  12.211297  5.271974  \n",
       "GSM2067435  12.129213  5.658377  \n",
       "GSM2067436  12.691983  4.899414  \n",
       "GSM2067437  12.032507  4.576603  \n",
       "GSM2067438  12.525255  5.089923  \n",
       "GSM2067439  11.593306  2.836373  \n",
       "GSM2067440  12.481041  3.461828  \n",
       "GSM2067441  13.381455  4.433457  \n",
       "GSM2067442  12.934966  5.089310  \n",
       "GSM2067443  12.242868  5.473535  \n",
       "GSM2067444  12.662624  5.340743  \n",
       "GSM2067445  13.345444  4.358263  \n",
       "GSM2067446  12.937860  4.567399  \n",
       "GSM2067447  12.877521  5.118452  \n",
       "GSM2067448  12.797197  5.146932  \n",
       "GSM2067449  12.584850  4.554525  \n",
       "GSM2067450  13.253766  5.249892  \n",
       "GSM2067451  12.833095  5.960049  \n",
       "GSM2067452  13.070649  6.031067  \n",
       "GSM2067453  13.338604  5.386587  \n",
       "GSM2067454  12.802153  5.418139  \n",
       "GSM2067455  12.110378  4.437177  \n",
       "GSM2067456  12.578838  5.433254  \n",
       "GSM2067457  12.872514  5.885852  \n",
       "GSM2067458  12.954499  4.147930  \n",
       "GSM2067459  12.288324  4.563690  \n",
       "GSM2067460  12.264850  4.904048  \n",
       "GSM2067461  13.421458  5.616683  \n",
       "GSM2067462  12.969342  5.282989  \n",
       "\n",
       "[60 rows x 14032 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 60 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1.0' with 30 occurrences. This represents 50.00% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 18.75\n",
      "  50% (Median): 27.0\n",
      "  75%: 35.0\n",
      "Min: 15.0\n",
      "Max: 60.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 20 occurrences. This represents 33.33% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression and response to psychological therapy\"\n",
      "!Series_summary\t\"This study represents the first investigation of genome-wide expression profiles with respect to psychological treatment outcome. Participants (n=102) with panic disorder or specific phobia received exposure-based CBT. Treatment outcome was defined as percentage reduction from baseline in clinician-rated severity of their primary anxiety diagnosis at post-treatment and six month follow-up. Gene expression was determined from whole blood samples at 3 time-points using the Illumina HT-12v4 BeadChip microarray. No changes in gene expression were significantly associated with treatment outcomes when correcting for multiple testing (q<0.05), although a small number of genes showed a suggestive association with treatment outcome (q<0.5, n=20). Study reports suggestive evidence for the role of a small number of genes in treatment outcome. Although preliminary, the findings contribute to a growing body of research suggesting that response to psychological therapies may be associated with changes at a biological level.\"\n",
      "!Series_overall_design\t\"Whole blood RNA was collected from patients (n=102) receiving exposure-based CBT at pre- and post-treatment and at follow-up, for investigation of association with therapy outcome. Includes 9 technical replicates.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2469746</th>\n",
       "      <th>GSM2469747</th>\n",
       "      <th>GSM2469748</th>\n",
       "      <th>GSM2469749</th>\n",
       "      <th>GSM2469750</th>\n",
       "      <th>GSM2469751</th>\n",
       "      <th>GSM2469752</th>\n",
       "      <th>GSM2469753</th>\n",
       "      <th>GSM2469754</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2470051</th>\n",
       "      <th>GSM2470052</th>\n",
       "      <th>GSM2470053</th>\n",
       "      <th>GSM2470054</th>\n",
       "      <th>GSM2470055</th>\n",
       "      <th>GSM2470056</th>\n",
       "      <th>GSM2470057</th>\n",
       "      <th>GSM2470058</th>\n",
       "      <th>GSM2470059</th>\n",
       "      <th>GSM2470060</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "      <td>gender: FEMALE</td>\n",
       "      <td>gender: MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "      <td>tissue: Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: post</td>\n",
       "      <td>timepoint: follow-up</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: post</td>\n",
       "      <td>timepoint: follow-up</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: post</td>\n",
       "      <td>timepoint: follow-up</td>\n",
       "      <td>...</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: post</td>\n",
       "      <td>timepoint: follow-up</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: follow-up</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: pre</td>\n",
       "      <td>timepoint: post</td>\n",
       "      <td>timepoint: follow-up</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 316 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession      GSM2469746       GSM2469747  \\\n",
       "0  !Sample_characteristics_ch1  gender: FEMALE   gender: FEMALE   \n",
       "1  !Sample_characteristics_ch1   tissue: Blood    tissue: Blood   \n",
       "2  !Sample_characteristics_ch1  timepoint: pre  timepoint: post   \n",
       "\n",
       "             GSM2469748      GSM2469749       GSM2469750  \\\n",
       "0        gender: FEMALE    gender: MALE     gender: MALE   \n",
       "1         tissue: Blood   tissue: Blood    tissue: Blood   \n",
       "2  timepoint: follow-up  timepoint: pre  timepoint: post   \n",
       "\n",
       "             GSM2469751      GSM2469752       GSM2469753  \\\n",
       "0          gender: MALE    gender: MALE     gender: MALE   \n",
       "1         tissue: Blood   tissue: Blood    tissue: Blood   \n",
       "2  timepoint: follow-up  timepoint: pre  timepoint: post   \n",
       "\n",
       "             GSM2469754  ...      GSM2470051       GSM2470052  \\\n",
       "0          gender: MALE  ...    gender: MALE     gender: MALE   \n",
       "1         tissue: Blood  ...   tissue: Blood    tissue: Blood   \n",
       "2  timepoint: follow-up  ...  timepoint: pre  timepoint: post   \n",
       "\n",
       "             GSM2470053      GSM2470054      GSM2470055            GSM2470056  \\\n",
       "0        gender: FEMALE  gender: FEMALE  gender: FEMALE          gender: MALE   \n",
       "1         tissue: Blood   tissue: Blood   tissue: Blood         tissue: Blood   \n",
       "2  timepoint: follow-up  timepoint: pre  timepoint: pre  timepoint: follow-up   \n",
       "\n",
       "       GSM2470057      GSM2470058       GSM2470059            GSM2470060  \n",
       "0  gender: FEMALE    gender: MALE   gender: FEMALE          gender: MALE  \n",
       "1   tissue: Blood   tissue: Blood    tissue: Blood         tissue: Blood  \n",
       "2  timepoint: pre  timepoint: pre  timepoint: post  timepoint: follow-up  \n",
       "\n",
       "[3 rows x 316 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No trait convert\n",
    "from utils import *\n",
    "cohort = accession_num = \"GSE94119\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Peripheral blood transcriptome profiles from an RNA Pilot Study within the United States Health and Retirement Study (HRS)\"\n",
      "!Series_summary\t\"Individual differences in peripheral blood transcriptomes in older adults as a function of demographic, socio-economic, psychological, and health history characteristics.\"\n",
      "!Series_overall_design\t\"Gene expression profiling was carried out on peripheral blood RNA samples collected from 121 community dwelling older adults participating in the United States Health and Retirement Study.  In addition to basic demographic characteristics (age, sex, race/ethnicity), participants were also assessed on health-related characteristics (body mass index/BMI; history of smoking or heavy alcohol consumption; history of chronic illnesses such as diabetes, cardiovascular disease, cancer, stroke), household annual income (log transformed), and measures of loneliness (UCLA Loneliness Scale; Russell D, Peplau LA, Cutrona CE: The revised UCLA Loneliness Scale: concurrent and discriminant validity evidence. J Pers Soc Psychol 1980, 39:472-480; higher values indicate greater loneliness) and purpose in life (PWB_Purpose; Purpose scale of the Ryff Scales of Psychological Well-being; Ryff CD (1989) Happiness is everything, or is it?  Explorations on the meaning of psychological well-being. J. Pers Soc. Psychol 57: 1069-1081; higher values indicate greater experience of purpose in life).  Additional analyses examined measures of optimism and pessimism (6 items from the Life Orientation Test; Scheier MF Carver CS Bridges MW (1994) Distinguishing optimism from neuroticism (and trait anxiety, self-mastery, and self-esteem): A reevaluation of the Life Orientation Test. J Pers Soc Psychol 67:1063-1078; higher values indicating greater optimism or pessimism), anxiety symptoms (5 items from Beck Anxiety Inventory; Beck AT et al (1988) An inventory for measuring clinical anxiety: Psychometric properties. J Consult Clin Psychol 56:893-897; higher values indicating greater anxietn), depressive symptoms (CESD; Center for Epidemiologic Studies Depression Scale; Radloff LS (1977) The CES-D scale:  A self-report depression scale for research in the general population. Applied Psychol Meas 1:386-401; higher values indicating greater depression), life satisfaction (Diener E et al (1985) The Satisfaction With Life Scale. J Pers Assess 49:71-75; higher values indicate greater life satisfaction), and positive affect (PA) and negative affect (NA; Watson D et al (1988) Development and validation of brief measures of positive and negative affect:  The PANAS scales. J Pers Soc Psychol 54:1063-1070; higher values indicate greater positive or negative affect).\"\n",
      "!Series_overall_design\t\"Binary characteristics are coded yes=1 and no=0.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1674313</th>\n",
       "      <th>GSM1674314</th>\n",
       "      <th>GSM1674315</th>\n",
       "      <th>GSM1674316</th>\n",
       "      <th>GSM1674317</th>\n",
       "      <th>GSM1674318</th>\n",
       "      <th>GSM1674319</th>\n",
       "      <th>GSM1674320</th>\n",
       "      <th>GSM1674321</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1674424</th>\n",
       "      <th>GSM1674425</th>\n",
       "      <th>GSM1674426</th>\n",
       "      <th>GSM1674427</th>\n",
       "      <th>GSM1674428</th>\n",
       "      <th>GSM1674429</th>\n",
       "      <th>GSM1674430</th>\n",
       "      <th>GSM1674431</th>\n",
       "      <th>GSM1674432</th>\n",
       "      <th>GSM1674433</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age (yrs): 79</td>\n",
       "      <td>age (yrs): 79</td>\n",
       "      <td>age (yrs): 76</td>\n",
       "      <td>age (yrs): 70</td>\n",
       "      <td>age (yrs): 65</td>\n",
       "      <td>age (yrs): 64</td>\n",
       "      <td>age (yrs): 75</td>\n",
       "      <td>age (yrs): 70</td>\n",
       "      <td>age (yrs): 66</td>\n",
       "      <td>...</td>\n",
       "      <td>age (yrs): 83</td>\n",
       "      <td>age (yrs): 89</td>\n",
       "      <td>age (yrs): 85</td>\n",
       "      <td>age (yrs): 88</td>\n",
       "      <td>age (yrs): 87</td>\n",
       "      <td>age (yrs): 81</td>\n",
       "      <td>age (yrs): 72</td>\n",
       "      <td>age (yrs): 66</td>\n",
       "      <td>age (yrs): 71</td>\n",
       "      <td>age (yrs): 73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 1</td>\n",
       "      <td>female: 0</td>\n",
       "      <td>female: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 1</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "      <td>black: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "      <td>hispanic: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>bmi: 22.7</td>\n",
       "      <td>bmi: 29.1</td>\n",
       "      <td>bmi: 25.8</td>\n",
       "      <td>bmi: 24.8</td>\n",
       "      <td>bmi: 42.1</td>\n",
       "      <td>bmi: 29.6</td>\n",
       "      <td>bmi: 21.4</td>\n",
       "      <td>bmi: 32.7</td>\n",
       "      <td>bmi: 30.7</td>\n",
       "      <td>...</td>\n",
       "      <td>bmi: 27.4</td>\n",
       "      <td>bmi: 23.6</td>\n",
       "      <td>bmi: 22.3</td>\n",
       "      <td>bmi: 18.6</td>\n",
       "      <td>bmi: 24.2</td>\n",
       "      <td>bmi: 24.5</td>\n",
       "      <td>bmi: 23.2</td>\n",
       "      <td>bmi: 24.4</td>\n",
       "      <td>bmi: 23.5</td>\n",
       "      <td>bmi: 29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 0</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 0</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 0</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 0</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 1</td>\n",
       "      <td>diabcvdcastr: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>ln_hh_income: 16.03</td>\n",
       "      <td>ln_hh_income: 15.49</td>\n",
       "      <td>ln_hh_income: 15.34</td>\n",
       "      <td>ln_hh_income: 15.52</td>\n",
       "      <td>ln_hh_income: 16.41</td>\n",
       "      <td>ln_hh_income: 15.34</td>\n",
       "      <td>ln_hh_income: 16.03</td>\n",
       "      <td>ln_hh_income: 14.20</td>\n",
       "      <td>ln_hh_income: 15.49</td>\n",
       "      <td>...</td>\n",
       "      <td>ln_hh_income: 14.03</td>\n",
       "      <td>ln_hh_income: 13.95</td>\n",
       "      <td>ln_hh_income: 13.96</td>\n",
       "      <td>ln_hh_income: 14.10</td>\n",
       "      <td>ln_hh_income: 13.96</td>\n",
       "      <td>ln_hh_income: 15.58</td>\n",
       "      <td>ln_hh_income: 14.35</td>\n",
       "      <td>ln_hh_income: 16.01</td>\n",
       "      <td>ln_hh_income: 17.90</td>\n",
       "      <td>ln_hh_income: 15.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 0</td>\n",
       "      <td>smoke: 0</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 0</td>\n",
       "      <td>smoke: 0</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 0</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "      <td>smoke: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 1</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: missing</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "      <td>alcohol: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 2.00</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.67</td>\n",
       "      <td>loneliness: 1.33</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>loneliness: 1.33</td>\n",
       "      <td>loneliness: 2.00</td>\n",
       "      <td>loneliness: 1.67</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.33</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.67</td>\n",
       "      <td>loneliness: 1.33</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "      <td>loneliness: 1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>pwb_purpose: 4.14</td>\n",
       "      <td>pwb_purpose: 6.00</td>\n",
       "      <td>pwb_purpose: 5.00</td>\n",
       "      <td>pwb_purpose: 5.43</td>\n",
       "      <td>pwb_purpose: 3.29</td>\n",
       "      <td>pwb_purpose: 5.00</td>\n",
       "      <td>pwb_purpose: 5.14</td>\n",
       "      <td>pwb_purpose: 4.71</td>\n",
       "      <td>pwb_purpose: 5.33</td>\n",
       "      <td>...</td>\n",
       "      <td>pwb_purpose: 6.00</td>\n",
       "      <td>pwb_purpose: 2.67</td>\n",
       "      <td>pwb_purpose: 3.50</td>\n",
       "      <td>pwb_purpose: 4.83</td>\n",
       "      <td>pwb_purpose: 5.50</td>\n",
       "      <td>pwb_purpose: 5.67</td>\n",
       "      <td>pwb_purpose: 2.71</td>\n",
       "      <td>pwb_purpose: 3.83</td>\n",
       "      <td>pwb_purpose: 6.00</td>\n",
       "      <td>pwb_purpose: 4.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>optimism: 4</td>\n",
       "      <td>optimism: 5.66667</td>\n",
       "      <td>optimism: 2.66667</td>\n",
       "      <td>optimism: 5.66667</td>\n",
       "      <td>optimism: 3</td>\n",
       "      <td>optimism: 4</td>\n",
       "      <td>optimism: 5</td>\n",
       "      <td>optimism: 1.66667</td>\n",
       "      <td>optimism: 6</td>\n",
       "      <td>...</td>\n",
       "      <td>optimism: 5.33333</td>\n",
       "      <td>optimism: missing</td>\n",
       "      <td>optimism: 4.66667</td>\n",
       "      <td>optimism: 3.66667</td>\n",
       "      <td>optimism: 5.33333</td>\n",
       "      <td>optimism: 5.66667</td>\n",
       "      <td>optimism: 5.33333</td>\n",
       "      <td>optimism: 3</td>\n",
       "      <td>optimism: 6</td>\n",
       "      <td>optimism: 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>pessimism: 1.66667</td>\n",
       "      <td>pessimism: 1</td>\n",
       "      <td>pessimism: 5</td>\n",
       "      <td>pessimism: 1</td>\n",
       "      <td>pessimism: 4</td>\n",
       "      <td>pessimism: 3</td>\n",
       "      <td>pessimism: 1</td>\n",
       "      <td>pessimism: 3.66667</td>\n",
       "      <td>pessimism: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>pessimism: 1</td>\n",
       "      <td>pessimism: missing</td>\n",
       "      <td>pessimism: 2</td>\n",
       "      <td>pessimism: 2.5</td>\n",
       "      <td>pessimism: 2</td>\n",
       "      <td>pessimism: 2</td>\n",
       "      <td>pessimism: 2.66667</td>\n",
       "      <td>pessimism: 3.66667</td>\n",
       "      <td>pessimism: 1</td>\n",
       "      <td>pessimism: 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1.8</td>\n",
       "      <td>anxiety: 1.2</td>\n",
       "      <td>anxiety: 1.4</td>\n",
       "      <td>anxiety: 1.2</td>\n",
       "      <td>anxiety: 1.2</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1.4</td>\n",
       "      <td>...</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: missing</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1.4</td>\n",
       "      <td>anxiety: 1.8</td>\n",
       "      <td>anxiety: 1</td>\n",
       "      <td>anxiety: 1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 1</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 2</td>\n",
       "      <td>cesd: 2</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: missing</td>\n",
       "      <td>cesd: 5</td>\n",
       "      <td>cesd: 1</td>\n",
       "      <td>cesd: 2</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 1</td>\n",
       "      <td>cesd: 4</td>\n",
       "      <td>cesd: 0</td>\n",
       "      <td>cesd: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>lifesatisfaction: 5.2</td>\n",
       "      <td>lifesatisfaction: 7</td>\n",
       "      <td>lifesatisfaction: 4.4</td>\n",
       "      <td>lifesatisfaction: 5.4</td>\n",
       "      <td>lifesatisfaction: 2.6</td>\n",
       "      <td>lifesatisfaction: 5.2</td>\n",
       "      <td>lifesatisfaction: 5.4</td>\n",
       "      <td>lifesatisfaction: 4.8</td>\n",
       "      <td>lifesatisfaction: 6.8</td>\n",
       "      <td>...</td>\n",
       "      <td>lifesatisfaction: 6.4</td>\n",
       "      <td>lifesatisfaction: missing</td>\n",
       "      <td>lifesatisfaction: 4</td>\n",
       "      <td>lifesatisfaction: 1</td>\n",
       "      <td>lifesatisfaction: 6.2</td>\n",
       "      <td>lifesatisfaction: 6.2</td>\n",
       "      <td>lifesatisfaction: 4.33333</td>\n",
       "      <td>lifesatisfaction: 2.8</td>\n",
       "      <td>lifesatisfaction: 6</td>\n",
       "      <td>lifesatisfaction: 5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>pa: 4</td>\n",
       "      <td>pa: 4.92308</td>\n",
       "      <td>pa: 2.84615</td>\n",
       "      <td>pa: 4.33333</td>\n",
       "      <td>pa: 2.66667</td>\n",
       "      <td>pa: 3.61538</td>\n",
       "      <td>pa: 4</td>\n",
       "      <td>pa: 4</td>\n",
       "      <td>pa: 4.92308</td>\n",
       "      <td>...</td>\n",
       "      <td>pa: 5</td>\n",
       "      <td>pa: missing</td>\n",
       "      <td>pa: missing</td>\n",
       "      <td>pa: 4.30769</td>\n",
       "      <td>pa: 4.53846</td>\n",
       "      <td>pa: 4</td>\n",
       "      <td>pa: 3</td>\n",
       "      <td>pa: 2.76923</td>\n",
       "      <td>pa: 5</td>\n",
       "      <td>pa: 4.30769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>na: 1.33333</td>\n",
       "      <td>na: 1</td>\n",
       "      <td>na: 1.33333</td>\n",
       "      <td>na: 1.33333</td>\n",
       "      <td>na: 1.33333</td>\n",
       "      <td>na: 1.58333</td>\n",
       "      <td>na: 1.33333</td>\n",
       "      <td>na: 1</td>\n",
       "      <td>na: 1.5</td>\n",
       "      <td>...</td>\n",
       "      <td>na: 1</td>\n",
       "      <td>na: missing</td>\n",
       "      <td>na: missing</td>\n",
       "      <td>na: 1.5</td>\n",
       "      <td>na: 1.33333</td>\n",
       "      <td>na: 1.16667</td>\n",
       "      <td>na: 2</td>\n",
       "      <td>na: 1.83333</td>\n",
       "      <td>na: 1</td>\n",
       "      <td>na: 1.41667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18 rows Ã— 122 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          !Sample_geo_accession             GSM1674313           GSM1674314  \\\n",
       "0   !Sample_characteristics_ch1          age (yrs): 79        age (yrs): 79   \n",
       "1   !Sample_characteristics_ch1              female: 0            female: 0   \n",
       "2   !Sample_characteristics_ch1               black: 0             black: 0   \n",
       "3   !Sample_characteristics_ch1            hispanic: 0          hispanic: 0   \n",
       "4   !Sample_characteristics_ch1              bmi: 22.7            bmi: 29.1   \n",
       "5   !Sample_characteristics_ch1        diabcvdcastr: 1      diabcvdcastr: 1   \n",
       "6   !Sample_characteristics_ch1    ln_hh_income: 16.03  ln_hh_income: 15.49   \n",
       "7   !Sample_characteristics_ch1               smoke: 1             smoke: 1   \n",
       "8   !Sample_characteristics_ch1             alcohol: 0           alcohol: 0   \n",
       "9   !Sample_characteristics_ch1       loneliness: 1.00     loneliness: 1.00   \n",
       "10  !Sample_characteristics_ch1      pwb_purpose: 4.14    pwb_purpose: 6.00   \n",
       "11  !Sample_characteristics_ch1            optimism: 4    optimism: 5.66667   \n",
       "12  !Sample_characteristics_ch1     pessimism: 1.66667         pessimism: 1   \n",
       "13  !Sample_characteristics_ch1             anxiety: 1           anxiety: 1   \n",
       "14  !Sample_characteristics_ch1                cesd: 0              cesd: 0   \n",
       "15  !Sample_characteristics_ch1  lifesatisfaction: 5.2  lifesatisfaction: 7   \n",
       "16  !Sample_characteristics_ch1                  pa: 4          pa: 4.92308   \n",
       "17  !Sample_characteristics_ch1            na: 1.33333                na: 1   \n",
       "\n",
       "               GSM1674315             GSM1674316             GSM1674317  \\\n",
       "0           age (yrs): 76          age (yrs): 70          age (yrs): 65   \n",
       "1               female: 0              female: 0              female: 1   \n",
       "2                black: 0               black: 0               black: 0   \n",
       "3             hispanic: 0            hispanic: 0            hispanic: 0   \n",
       "4               bmi: 25.8              bmi: 24.8              bmi: 42.1   \n",
       "5         diabcvdcastr: 1        diabcvdcastr: 0        diabcvdcastr: 1   \n",
       "6     ln_hh_income: 15.34    ln_hh_income: 15.52    ln_hh_income: 16.41   \n",
       "7                smoke: 1               smoke: 1               smoke: 0   \n",
       "8              alcohol: 0             alcohol: 1             alcohol: 0   \n",
       "9        loneliness: 2.00       loneliness: 1.00       loneliness: 1.67   \n",
       "10      pwb_purpose: 5.00      pwb_purpose: 5.43      pwb_purpose: 3.29   \n",
       "11      optimism: 2.66667      optimism: 5.66667            optimism: 3   \n",
       "12           pessimism: 5           pessimism: 1           pessimism: 4   \n",
       "13           anxiety: 1.8           anxiety: 1.2           anxiety: 1.4   \n",
       "14                cesd: 1                cesd: 0                cesd: 2   \n",
       "15  lifesatisfaction: 4.4  lifesatisfaction: 5.4  lifesatisfaction: 2.6   \n",
       "16            pa: 2.84615            pa: 4.33333            pa: 2.66667   \n",
       "17            na: 1.33333            na: 1.33333            na: 1.33333   \n",
       "\n",
       "               GSM1674318             GSM1674319             GSM1674320  \\\n",
       "0           age (yrs): 64          age (yrs): 75          age (yrs): 70   \n",
       "1               female: 1              female: 1              female: 0   \n",
       "2                black: 0               black: 0               black: 0   \n",
       "3             hispanic: 0            hispanic: 0            hispanic: 0   \n",
       "4               bmi: 29.6              bmi: 21.4              bmi: 32.7   \n",
       "5         diabcvdcastr: 1        diabcvdcastr: 0        diabcvdcastr: 1   \n",
       "6     ln_hh_income: 15.34    ln_hh_income: 16.03    ln_hh_income: 14.20   \n",
       "7                smoke: 0               smoke: 1               smoke: 1   \n",
       "8              alcohol: 0             alcohol: 0             alcohol: 0   \n",
       "9        loneliness: 1.33       loneliness: 1.00       loneliness: 1.00   \n",
       "10      pwb_purpose: 5.00      pwb_purpose: 5.14      pwb_purpose: 4.71   \n",
       "11            optimism: 4            optimism: 5      optimism: 1.66667   \n",
       "12           pessimism: 3           pessimism: 1     pessimism: 3.66667   \n",
       "13           anxiety: 1.2           anxiety: 1.2             anxiety: 1   \n",
       "14                cesd: 2                cesd: 0                cesd: 0   \n",
       "15  lifesatisfaction: 5.2  lifesatisfaction: 5.4  lifesatisfaction: 4.8   \n",
       "16            pa: 3.61538                  pa: 4                  pa: 4   \n",
       "17            na: 1.58333            na: 1.33333                  na: 1   \n",
       "\n",
       "               GSM1674321  ...             GSM1674424  \\\n",
       "0           age (yrs): 66  ...          age (yrs): 83   \n",
       "1               female: 1  ...              female: 1   \n",
       "2                black: 0  ...               black: 0   \n",
       "3             hispanic: 0  ...            hispanic: 0   \n",
       "4               bmi: 30.7  ...              bmi: 27.4   \n",
       "5         diabcvdcastr: 0  ...        diabcvdcastr: 1   \n",
       "6     ln_hh_income: 15.49  ...    ln_hh_income: 14.03   \n",
       "7                smoke: 1  ...               smoke: 1   \n",
       "8              alcohol: 0  ...             alcohol: 0   \n",
       "9        loneliness: 1.00  ...       loneliness: 1.33   \n",
       "10      pwb_purpose: 5.33  ...      pwb_purpose: 6.00   \n",
       "11            optimism: 6  ...      optimism: 5.33333   \n",
       "12           pessimism: 1  ...           pessimism: 1   \n",
       "13           anxiety: 1.4  ...             anxiety: 1   \n",
       "14                cesd: 0  ...                cesd: 0   \n",
       "15  lifesatisfaction: 6.8  ...  lifesatisfaction: 6.4   \n",
       "16            pa: 4.92308  ...                  pa: 5   \n",
       "17                na: 1.5  ...                  na: 1   \n",
       "\n",
       "                   GSM1674425           GSM1674426           GSM1674427  \\\n",
       "0               age (yrs): 89        age (yrs): 85        age (yrs): 88   \n",
       "1                   female: 0            female: 1            female: 1   \n",
       "2                    black: 0             black: 0             black: 0   \n",
       "3                 hispanic: 0          hispanic: 0          hispanic: 0   \n",
       "4                   bmi: 23.6            bmi: 22.3            bmi: 18.6   \n",
       "5             diabcvdcastr: 1      diabcvdcastr: 1      diabcvdcastr: 1   \n",
       "6         ln_hh_income: 13.95  ln_hh_income: 13.96  ln_hh_income: 14.10   \n",
       "7                    smoke: 1             smoke: 0             smoke: 0   \n",
       "8            alcohol: missing           alcohol: 0           alcohol: 0   \n",
       "9            loneliness: 2.00     loneliness: 1.67     loneliness: 1.00   \n",
       "10          pwb_purpose: 2.67    pwb_purpose: 3.50    pwb_purpose: 4.83   \n",
       "11          optimism: missing    optimism: 4.66667    optimism: 3.66667   \n",
       "12         pessimism: missing         pessimism: 2       pessimism: 2.5   \n",
       "13           anxiety: missing           anxiety: 1           anxiety: 1   \n",
       "14              cesd: missing              cesd: 5              cesd: 1   \n",
       "15  lifesatisfaction: missing  lifesatisfaction: 4  lifesatisfaction: 1   \n",
       "16                pa: missing          pa: missing          pa: 4.30769   \n",
       "17                na: missing          na: missing              na: 1.5   \n",
       "\n",
       "               GSM1674428             GSM1674429                 GSM1674430  \\\n",
       "0           age (yrs): 87          age (yrs): 81              age (yrs): 72   \n",
       "1               female: 0              female: 0                  female: 1   \n",
       "2                black: 0               black: 0                   black: 1   \n",
       "3             hispanic: 0            hispanic: 0                hispanic: 0   \n",
       "4               bmi: 24.2              bmi: 24.5                  bmi: 23.2   \n",
       "5         diabcvdcastr: 0        diabcvdcastr: 1            diabcvdcastr: 0   \n",
       "6     ln_hh_income: 13.96    ln_hh_income: 15.58        ln_hh_income: 14.35   \n",
       "7                smoke: 1               smoke: 1                   smoke: 0   \n",
       "8              alcohol: 0             alcohol: 0                 alcohol: 0   \n",
       "9        loneliness: 1.33       loneliness: 1.00           loneliness: 1.67   \n",
       "10      pwb_purpose: 5.50      pwb_purpose: 5.67          pwb_purpose: 2.71   \n",
       "11      optimism: 5.33333      optimism: 5.66667          optimism: 5.33333   \n",
       "12           pessimism: 2           pessimism: 2         pessimism: 2.66667   \n",
       "13             anxiety: 1             anxiety: 1               anxiety: 1.4   \n",
       "14                cesd: 2                cesd: 0                    cesd: 1   \n",
       "15  lifesatisfaction: 6.2  lifesatisfaction: 6.2  lifesatisfaction: 4.33333   \n",
       "16            pa: 4.53846                  pa: 4                      pa: 3   \n",
       "17            na: 1.33333            na: 1.16667                      na: 2   \n",
       "\n",
       "               GSM1674431           GSM1674432             GSM1674433  \n",
       "0           age (yrs): 66        age (yrs): 71          age (yrs): 73  \n",
       "1               female: 1            female: 0              female: 0  \n",
       "2                black: 0             black: 0               black: 0  \n",
       "3             hispanic: 0          hispanic: 0            hispanic: 0  \n",
       "4               bmi: 24.4            bmi: 23.5              bmi: 29.0  \n",
       "5         diabcvdcastr: 1      diabcvdcastr: 1        diabcvdcastr: 0  \n",
       "6     ln_hh_income: 16.01  ln_hh_income: 17.90    ln_hh_income: 15.93  \n",
       "7                smoke: 1             smoke: 1               smoke: 1  \n",
       "8              alcohol: 0           alcohol: 0             alcohol: 0  \n",
       "9        loneliness: 1.33     loneliness: 1.00       loneliness: 1.00  \n",
       "10      pwb_purpose: 3.83    pwb_purpose: 6.00      pwb_purpose: 4.33  \n",
       "11            optimism: 3          optimism: 6            optimism: 5  \n",
       "12     pessimism: 3.66667         pessimism: 1           pessimism: 2  \n",
       "13           anxiety: 1.8           anxiety: 1           anxiety: 1.4  \n",
       "14                cesd: 4              cesd: 0                cesd: 0  \n",
       "15  lifesatisfaction: 2.8  lifesatisfaction: 6  lifesatisfaction: 5.6  \n",
       "16            pa: 2.76923                pa: 5            pa: 4.30769  \n",
       "17            na: 1.83333                na: 1            na: 1.41667  \n",
       "\n",
       "[18 rows x 122 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = accession_num = \"GSE68526\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression analysis in whole blood samples obtained from donors diagnosed with MDD compared to healthy controls\"\n",
      "!Series_summary\t\"This study investigated gene expression changes in whole blood samples obtained from donors diagnosed with major depressive disorder (MDD) compared to healthy controls. Micro-array data were available from whole blood on patients with MDD (N=128, 64 with generalised anxiety disorder, diagnosed by the MINI questionnaire, and 64 without anxiety disorder) and healthy controls (N=64). RNA was isolated from all samples using the standard PAXgene protocol on the Qiagen Biorobot 8000. All samples gave good quality RNA, as assessed by Agilent Bioanalyser. The yield range was 0.86-15.05ug with an average of 6.25ug. Samples were then randomised into batches, with each batch containing a representative number of controls, depression with anxiety and depression without anxiety, and the same ratio of females to males (3:1). 50ng of RNA from each sample was converted to a biotin labeled cDNA probe using NuGEN SPIA amplification. The probes were then hybridized to Affymetrix U133_Plus2.0 Genechips.\"\n",
      "!Series_overall_design\t\"Measurements from whole blood samples for 128 patients with MDD (among which 64 have been diagnosed with generalised anxiety disorder) and 64 healthy controls\"\n",
      "!Series_overall_design\t\"contributor: MRC ImmunoPsychiatry Consortium/GlaxoSmithKline\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2612096</th>\n",
       "      <th>GSM2612097</th>\n",
       "      <th>GSM2612098</th>\n",
       "      <th>GSM2612099</th>\n",
       "      <th>GSM2612100</th>\n",
       "      <th>GSM2612101</th>\n",
       "      <th>GSM2612102</th>\n",
       "      <th>GSM2612103</th>\n",
       "      <th>GSM2612104</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2612278</th>\n",
       "      <th>GSM2612279</th>\n",
       "      <th>GSM2612280</th>\n",
       "      <th>GSM2612281</th>\n",
       "      <th>GSM2612282</th>\n",
       "      <th>GSM2612283</th>\n",
       "      <th>GSM2612284</th>\n",
       "      <th>GSM2612285</th>\n",
       "      <th>GSM2612286</th>\n",
       "      <th>GSM2612287</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>...</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CASE; major depressive disorder...</td>\n",
       "      <td>subject group: CASE; major depressive disorder...</td>\n",
       "      <td>subject group: CASE; major depressive disorder...</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CASE; major depressive disorder...</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "      <td>subject group: CNTL; healthy control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>...</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: yes</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: yes</td>\n",
       "      <td>anxiety: no</td>\n",
       "      <td>anxiety: no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 35.8</td>\n",
       "      <td>age: 36.9</td>\n",
       "      <td>age: 62</td>\n",
       "      <td>age: 35.7</td>\n",
       "      <td>age: 53.4</td>\n",
       "      <td>age: 52.5</td>\n",
       "      <td>age: 55.2</td>\n",
       "      <td>age: 64.5</td>\n",
       "      <td>age: 52.9</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 66.7</td>\n",
       "      <td>age: 52.6</td>\n",
       "      <td>age: 62.2</td>\n",
       "      <td>age: 48.6</td>\n",
       "      <td>age: 32</td>\n",
       "      <td>age: 62.3</td>\n",
       "      <td>age: 73.1</td>\n",
       "      <td>age: 61.3</td>\n",
       "      <td>age: 31.2</td>\n",
       "      <td>age: 61.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "      <td>tissue: whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>batch: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "      <td>batch: 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 193 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                            GSM2612096  \\\n",
       "0  !Sample_characteristics_ch1  subject group: CNTL; healthy control   \n",
       "1  !Sample_characteristics_ch1                           anxiety: no   \n",
       "2  !Sample_characteristics_ch1                             gender: M   \n",
       "3  !Sample_characteristics_ch1                             age: 35.8   \n",
       "4  !Sample_characteristics_ch1                   tissue: whole blood   \n",
       "5  !Sample_characteristics_ch1                              batch: 1   \n",
       "\n",
       "                             GSM2612097                            GSM2612098  \\\n",
       "0  subject group: CNTL; healthy control  subject group: CNTL; healthy control   \n",
       "1                           anxiety: no                           anxiety: no   \n",
       "2                             gender: F                             gender: F   \n",
       "3                             age: 36.9                               age: 62   \n",
       "4                   tissue: whole blood                   tissue: whole blood   \n",
       "5                              batch: 1                              batch: 1   \n",
       "\n",
       "                             GSM2612099                            GSM2612100  \\\n",
       "0  subject group: CNTL; healthy control  subject group: CNTL; healthy control   \n",
       "1                           anxiety: no                           anxiety: no   \n",
       "2                             gender: F                             gender: M   \n",
       "3                             age: 35.7                             age: 53.4   \n",
       "4                   tissue: whole blood                   tissue: whole blood   \n",
       "5                              batch: 1                              batch: 1   \n",
       "\n",
       "                             GSM2612101                            GSM2612102  \\\n",
       "0  subject group: CNTL; healthy control  subject group: CNTL; healthy control   \n",
       "1                           anxiety: no                           anxiety: no   \n",
       "2                             gender: F                             gender: F   \n",
       "3                             age: 52.5                             age: 55.2   \n",
       "4                   tissue: whole blood                   tissue: whole blood   \n",
       "5                              batch: 1                              batch: 1   \n",
       "\n",
       "                             GSM2612103                            GSM2612104  \\\n",
       "0  subject group: CNTL; healthy control  subject group: CNTL; healthy control   \n",
       "1                           anxiety: no                           anxiety: no   \n",
       "2                             gender: F                             gender: F   \n",
       "3                             age: 64.5                             age: 52.9   \n",
       "4                   tissue: whole blood                   tissue: whole blood   \n",
       "5                              batch: 1                              batch: 1   \n",
       "\n",
       "   ...                            GSM2612278  \\\n",
       "0  ...  subject group: CNTL; healthy control   \n",
       "1  ...                           anxiety: no   \n",
       "2  ...                             gender: F   \n",
       "3  ...                             age: 66.7   \n",
       "4  ...                   tissue: whole blood   \n",
       "5  ...                              batch: 2   \n",
       "\n",
       "                             GSM2612279                            GSM2612280  \\\n",
       "0  subject group: CNTL; healthy control  subject group: CNTL; healthy control   \n",
       "1                           anxiety: no                           anxiety: no   \n",
       "2                             gender: F                             gender: F   \n",
       "3                             age: 52.6                             age: 62.2   \n",
       "4                   tissue: whole blood                   tissue: whole blood   \n",
       "5                              batch: 2                              batch: 2   \n",
       "\n",
       "                                          GSM2612281  \\\n",
       "0  subject group: CASE; major depressive disorder...   \n",
       "1                                       anxiety: yes   \n",
       "2                                          gender: F   \n",
       "3                                          age: 48.6   \n",
       "4                                tissue: whole blood   \n",
       "5                                           batch: 2   \n",
       "\n",
       "                                          GSM2612282  \\\n",
       "0  subject group: CASE; major depressive disorder...   \n",
       "1                                        anxiety: no   \n",
       "2                                          gender: F   \n",
       "3                                            age: 32   \n",
       "4                                tissue: whole blood   \n",
       "5                                           batch: 2   \n",
       "\n",
       "                                          GSM2612283  \\\n",
       "0  subject group: CASE; major depressive disorder...   \n",
       "1                                        anxiety: no   \n",
       "2                                          gender: F   \n",
       "3                                          age: 62.3   \n",
       "4                                tissue: whole blood   \n",
       "5                                           batch: 2   \n",
       "\n",
       "                             GSM2612284  \\\n",
       "0  subject group: CNTL; healthy control   \n",
       "1                           anxiety: no   \n",
       "2                             gender: F   \n",
       "3                             age: 73.1   \n",
       "4                   tissue: whole blood   \n",
       "5                              batch: 2   \n",
       "\n",
       "                                          GSM2612285  \\\n",
       "0  subject group: CASE; major depressive disorder...   \n",
       "1                                       anxiety: yes   \n",
       "2                                          gender: F   \n",
       "3                                          age: 61.3   \n",
       "4                                tissue: whole blood   \n",
       "5                                           batch: 2   \n",
       "\n",
       "                             GSM2612286                            GSM2612287  \n",
       "0  subject group: CNTL; healthy control  subject group: CNTL; healthy control  \n",
       "1                           anxiety: no                           anxiety: no  \n",
       "2                             gender: F                             gender: F  \n",
       "3                             age: 31.2                             age: 61.2  \n",
       "4                   tissue: whole blood                   tissue: whole blood  \n",
       "5                              batch: 2                              batch: 2  \n",
       "\n",
       "[6 rows x 193 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE98793\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'anxiety: no', 'anxiety: yes'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 3\n",
    "gender_row = 2\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'anxiety: yes':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = float(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "    \n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2612096</th>\n",
       "      <th>GSM2612097</th>\n",
       "      <th>GSM2612098</th>\n",
       "      <th>GSM2612099</th>\n",
       "      <th>GSM2612100</th>\n",
       "      <th>GSM2612101</th>\n",
       "      <th>GSM2612102</th>\n",
       "      <th>GSM2612103</th>\n",
       "      <th>GSM2612104</th>\n",
       "      <th>GSM2612105</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2612278</th>\n",
       "      <th>GSM2612279</th>\n",
       "      <th>GSM2612280</th>\n",
       "      <th>GSM2612281</th>\n",
       "      <th>GSM2612282</th>\n",
       "      <th>GSM2612283</th>\n",
       "      <th>GSM2612284</th>\n",
       "      <th>GSM2612285</th>\n",
       "      <th>GSM2612286</th>\n",
       "      <th>GSM2612287</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>35.8</td>\n",
       "      <td>36.9</td>\n",
       "      <td>62.0</td>\n",
       "      <td>35.7</td>\n",
       "      <td>53.4</td>\n",
       "      <td>52.5</td>\n",
       "      <td>55.2</td>\n",
       "      <td>64.5</td>\n",
       "      <td>52.9</td>\n",
       "      <td>69.6</td>\n",
       "      <td>...</td>\n",
       "      <td>66.7</td>\n",
       "      <td>52.6</td>\n",
       "      <td>62.2</td>\n",
       "      <td>48.6</td>\n",
       "      <td>32.0</td>\n",
       "      <td>62.3</td>\n",
       "      <td>73.1</td>\n",
       "      <td>61.3</td>\n",
       "      <td>31.2</td>\n",
       "      <td>61.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM2612096  GSM2612097  GSM2612098  GSM2612099  GSM2612100  \\\n",
       "Anxiety disorder         0.0         0.0         0.0         0.0         0.0   \n",
       "Age                     35.8        36.9        62.0        35.7        53.4   \n",
       "Gender                   0.0         1.0         1.0         1.0         0.0   \n",
       "\n",
       "                  GSM2612101  GSM2612102  GSM2612103  GSM2612104  GSM2612105  \\\n",
       "Anxiety disorder         0.0         0.0         0.0         0.0         0.0   \n",
       "Age                     52.5        55.2        64.5        52.9        69.6   \n",
       "Gender                   1.0         1.0         1.0         1.0         1.0   \n",
       "\n",
       "                  ...  GSM2612278  GSM2612279  GSM2612280  GSM2612281  \\\n",
       "Anxiety disorder  ...         0.0         0.0         0.0         1.0   \n",
       "Age               ...        66.7        52.6        62.2        48.6   \n",
       "Gender            ...         1.0         1.0         1.0         1.0   \n",
       "\n",
       "                  GSM2612282  GSM2612283  GSM2612284  GSM2612285  GSM2612286  \\\n",
       "Anxiety disorder         0.0         0.0         0.0         1.0         0.0   \n",
       "Age                     32.0        62.3        73.1        61.3        31.2   \n",
       "Gender                   1.0         1.0         1.0         1.0         1.0   \n",
       "\n",
       "                  GSM2612287  \n",
       "Anxiety disorder         0.0  \n",
       "Age                     61.2  \n",
       "Gender                   1.0  \n",
       "\n",
       "[3 rows x 192 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2612096</th>\n",
       "      <th>GSM2612097</th>\n",
       "      <th>GSM2612098</th>\n",
       "      <th>GSM2612099</th>\n",
       "      <th>GSM2612100</th>\n",
       "      <th>GSM2612101</th>\n",
       "      <th>GSM2612102</th>\n",
       "      <th>GSM2612103</th>\n",
       "      <th>GSM2612104</th>\n",
       "      <th>GSM2612105</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2612278</th>\n",
       "      <th>GSM2612279</th>\n",
       "      <th>GSM2612280</th>\n",
       "      <th>GSM2612281</th>\n",
       "      <th>GSM2612282</th>\n",
       "      <th>GSM2612283</th>\n",
       "      <th>GSM2612284</th>\n",
       "      <th>GSM2612285</th>\n",
       "      <th>GSM2612286</th>\n",
       "      <th>GSM2612287</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>5.850755</td>\n",
       "      <td>5.577231</td>\n",
       "      <td>5.663056</td>\n",
       "      <td>5.596154</td>\n",
       "      <td>5.242699</td>\n",
       "      <td>5.212191</td>\n",
       "      <td>4.872792</td>\n",
       "      <td>5.272285</td>\n",
       "      <td>5.338772</td>\n",
       "      <td>5.070973</td>\n",
       "      <td>...</td>\n",
       "      <td>6.414110</td>\n",
       "      <td>6.115670</td>\n",
       "      <td>5.679880</td>\n",
       "      <td>6.740236</td>\n",
       "      <td>5.972264</td>\n",
       "      <td>5.808947</td>\n",
       "      <td>5.867953</td>\n",
       "      <td>5.938182</td>\n",
       "      <td>6.000403</td>\n",
       "      <td>5.772250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>7.092003</td>\n",
       "      <td>6.618856</td>\n",
       "      <td>6.487570</td>\n",
       "      <td>6.565388</td>\n",
       "      <td>6.531346</td>\n",
       "      <td>6.073299</td>\n",
       "      <td>6.414285</td>\n",
       "      <td>6.099957</td>\n",
       "      <td>6.969604</td>\n",
       "      <td>6.837721</td>\n",
       "      <td>...</td>\n",
       "      <td>7.798097</td>\n",
       "      <td>7.802633</td>\n",
       "      <td>7.345928</td>\n",
       "      <td>7.758726</td>\n",
       "      <td>7.580560</td>\n",
       "      <td>7.444604</td>\n",
       "      <td>7.238209</td>\n",
       "      <td>7.513523</td>\n",
       "      <td>7.737975</td>\n",
       "      <td>7.702891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>9.373934</td>\n",
       "      <td>9.315652</td>\n",
       "      <td>8.237757</td>\n",
       "      <td>8.877479</td>\n",
       "      <td>8.148008</td>\n",
       "      <td>8.805715</td>\n",
       "      <td>8.859519</td>\n",
       "      <td>7.861680</td>\n",
       "      <td>9.129116</td>\n",
       "      <td>8.725871</td>\n",
       "      <td>...</td>\n",
       "      <td>9.967406</td>\n",
       "      <td>10.193932</td>\n",
       "      <td>9.737181</td>\n",
       "      <td>10.029855</td>\n",
       "      <td>9.960287</td>\n",
       "      <td>10.432011</td>\n",
       "      <td>10.155011</td>\n",
       "      <td>9.821582</td>\n",
       "      <td>9.522425</td>\n",
       "      <td>10.838637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>5.814709</td>\n",
       "      <td>5.643282</td>\n",
       "      <td>5.363979</td>\n",
       "      <td>5.340978</td>\n",
       "      <td>5.701092</td>\n",
       "      <td>5.293687</td>\n",
       "      <td>5.379093</td>\n",
       "      <td>5.975036</td>\n",
       "      <td>5.661394</td>\n",
       "      <td>5.439686</td>\n",
       "      <td>...</td>\n",
       "      <td>5.572871</td>\n",
       "      <td>5.547584</td>\n",
       "      <td>5.748475</td>\n",
       "      <td>5.630231</td>\n",
       "      <td>5.858684</td>\n",
       "      <td>5.945134</td>\n",
       "      <td>5.631375</td>\n",
       "      <td>5.695992</td>\n",
       "      <td>5.706208</td>\n",
       "      <td>6.016455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>2.728267</td>\n",
       "      <td>2.671652</td>\n",
       "      <td>2.206741</td>\n",
       "      <td>2.998085</td>\n",
       "      <td>2.594565</td>\n",
       "      <td>2.340774</td>\n",
       "      <td>2.754393</td>\n",
       "      <td>2.607237</td>\n",
       "      <td>2.532411</td>\n",
       "      <td>3.130581</td>\n",
       "      <td>...</td>\n",
       "      <td>2.157327</td>\n",
       "      <td>2.264339</td>\n",
       "      <td>2.443638</td>\n",
       "      <td>2.302708</td>\n",
       "      <td>2.523345</td>\n",
       "      <td>2.454866</td>\n",
       "      <td>2.416422</td>\n",
       "      <td>2.298950</td>\n",
       "      <td>2.235815</td>\n",
       "      <td>2.475537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>2.413047</td>\n",
       "      <td>2.499112</td>\n",
       "      <td>2.493270</td>\n",
       "      <td>2.176757</td>\n",
       "      <td>2.497286</td>\n",
       "      <td>2.473523</td>\n",
       "      <td>2.415532</td>\n",
       "      <td>2.959415</td>\n",
       "      <td>2.552571</td>\n",
       "      <td>2.484300</td>\n",
       "      <td>...</td>\n",
       "      <td>2.421918</td>\n",
       "      <td>2.395902</td>\n",
       "      <td>2.259330</td>\n",
       "      <td>2.362848</td>\n",
       "      <td>2.301254</td>\n",
       "      <td>2.554583</td>\n",
       "      <td>2.382011</td>\n",
       "      <td>2.411032</td>\n",
       "      <td>2.385052</td>\n",
       "      <td>2.573699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>3.248725</td>\n",
       "      <td>2.952288</td>\n",
       "      <td>3.036010</td>\n",
       "      <td>3.018452</td>\n",
       "      <td>2.857945</td>\n",
       "      <td>2.867773</td>\n",
       "      <td>2.934452</td>\n",
       "      <td>2.851396</td>\n",
       "      <td>3.078947</td>\n",
       "      <td>2.719665</td>\n",
       "      <td>...</td>\n",
       "      <td>2.702217</td>\n",
       "      <td>2.721138</td>\n",
       "      <td>2.791224</td>\n",
       "      <td>2.854180</td>\n",
       "      <td>2.759138</td>\n",
       "      <td>3.055240</td>\n",
       "      <td>2.988927</td>\n",
       "      <td>2.825169</td>\n",
       "      <td>2.891791</td>\n",
       "      <td>3.178614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>2.054429</td>\n",
       "      <td>2.049808</td>\n",
       "      <td>1.899813</td>\n",
       "      <td>2.431576</td>\n",
       "      <td>2.212574</td>\n",
       "      <td>2.339308</td>\n",
       "      <td>2.407863</td>\n",
       "      <td>1.997606</td>\n",
       "      <td>2.412021</td>\n",
       "      <td>2.173443</td>\n",
       "      <td>...</td>\n",
       "      <td>1.929942</td>\n",
       "      <td>1.947370</td>\n",
       "      <td>1.879175</td>\n",
       "      <td>2.075299</td>\n",
       "      <td>1.967994</td>\n",
       "      <td>2.045717</td>\n",
       "      <td>1.961351</td>\n",
       "      <td>2.030025</td>\n",
       "      <td>1.918269</td>\n",
       "      <td>2.202680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>3.187857</td>\n",
       "      <td>2.929797</td>\n",
       "      <td>2.686815</td>\n",
       "      <td>2.938610</td>\n",
       "      <td>2.973220</td>\n",
       "      <td>2.737518</td>\n",
       "      <td>2.637788</td>\n",
       "      <td>2.782907</td>\n",
       "      <td>3.149515</td>\n",
       "      <td>2.969094</td>\n",
       "      <td>...</td>\n",
       "      <td>2.638295</td>\n",
       "      <td>2.821802</td>\n",
       "      <td>2.744246</td>\n",
       "      <td>2.854341</td>\n",
       "      <td>2.791133</td>\n",
       "      <td>3.237114</td>\n",
       "      <td>2.943271</td>\n",
       "      <td>2.863169</td>\n",
       "      <td>2.741495</td>\n",
       "      <td>3.075356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>3.172204</td>\n",
       "      <td>2.850546</td>\n",
       "      <td>3.098309</td>\n",
       "      <td>2.796332</td>\n",
       "      <td>3.087662</td>\n",
       "      <td>2.914371</td>\n",
       "      <td>2.666093</td>\n",
       "      <td>3.219344</td>\n",
       "      <td>2.917216</td>\n",
       "      <td>3.096693</td>\n",
       "      <td>...</td>\n",
       "      <td>2.978033</td>\n",
       "      <td>2.983088</td>\n",
       "      <td>3.023049</td>\n",
       "      <td>3.055407</td>\n",
       "      <td>3.099678</td>\n",
       "      <td>3.405567</td>\n",
       "      <td>3.172511</td>\n",
       "      <td>3.002204</td>\n",
       "      <td>3.038403</td>\n",
       "      <td>3.359097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54675 rows Ã— 192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM2612096  GSM2612097  GSM2612098  GSM2612099  GSM2612100  \\\n",
       "ID                                                                            \n",
       "1007_s_at          5.850755    5.577231    5.663056    5.596154    5.242699   \n",
       "1053_at            7.092003    6.618856    6.487570    6.565388    6.531346   \n",
       "117_at             9.373934    9.315652    8.237757    8.877479    8.148008   \n",
       "121_at             5.814709    5.643282    5.363979    5.340978    5.701092   \n",
       "1255_g_at          2.728267    2.671652    2.206741    2.998085    2.594565   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     2.413047    2.499112    2.493270    2.176757    2.497286   \n",
       "AFFX-ThrX-M_at     3.248725    2.952288    3.036010    3.018452    2.857945   \n",
       "AFFX-TrpnX-3_at    2.054429    2.049808    1.899813    2.431576    2.212574   \n",
       "AFFX-TrpnX-5_at    3.187857    2.929797    2.686815    2.938610    2.973220   \n",
       "AFFX-TrpnX-M_at    3.172204    2.850546    3.098309    2.796332    3.087662   \n",
       "\n",
       "                 GSM2612101  GSM2612102  GSM2612103  GSM2612104  GSM2612105  \\\n",
       "ID                                                                            \n",
       "1007_s_at          5.212191    4.872792    5.272285    5.338772    5.070973   \n",
       "1053_at            6.073299    6.414285    6.099957    6.969604    6.837721   \n",
       "117_at             8.805715    8.859519    7.861680    9.129116    8.725871   \n",
       "121_at             5.293687    5.379093    5.975036    5.661394    5.439686   \n",
       "1255_g_at          2.340774    2.754393    2.607237    2.532411    3.130581   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     2.473523    2.415532    2.959415    2.552571    2.484300   \n",
       "AFFX-ThrX-M_at     2.867773    2.934452    2.851396    3.078947    2.719665   \n",
       "AFFX-TrpnX-3_at    2.339308    2.407863    1.997606    2.412021    2.173443   \n",
       "AFFX-TrpnX-5_at    2.737518    2.637788    2.782907    3.149515    2.969094   \n",
       "AFFX-TrpnX-M_at    2.914371    2.666093    3.219344    2.917216    3.096693   \n",
       "\n",
       "                 ...  GSM2612278  GSM2612279  GSM2612280  GSM2612281  \\\n",
       "ID               ...                                                   \n",
       "1007_s_at        ...    6.414110    6.115670    5.679880    6.740236   \n",
       "1053_at          ...    7.798097    7.802633    7.345928    7.758726   \n",
       "117_at           ...    9.967406   10.193932    9.737181   10.029855   \n",
       "121_at           ...    5.572871    5.547584    5.748475    5.630231   \n",
       "1255_g_at        ...    2.157327    2.264339    2.443638    2.302708   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at   ...    2.421918    2.395902    2.259330    2.362848   \n",
       "AFFX-ThrX-M_at   ...    2.702217    2.721138    2.791224    2.854180   \n",
       "AFFX-TrpnX-3_at  ...    1.929942    1.947370    1.879175    2.075299   \n",
       "AFFX-TrpnX-5_at  ...    2.638295    2.821802    2.744246    2.854341   \n",
       "AFFX-TrpnX-M_at  ...    2.978033    2.983088    3.023049    3.055407   \n",
       "\n",
       "                 GSM2612282  GSM2612283  GSM2612284  GSM2612285  GSM2612286  \\\n",
       "ID                                                                            \n",
       "1007_s_at          5.972264    5.808947    5.867953    5.938182    6.000403   \n",
       "1053_at            7.580560    7.444604    7.238209    7.513523    7.737975   \n",
       "117_at             9.960287   10.432011   10.155011    9.821582    9.522425   \n",
       "121_at             5.858684    5.945134    5.631375    5.695992    5.706208   \n",
       "1255_g_at          2.523345    2.454866    2.416422    2.298950    2.235815   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     2.301254    2.554583    2.382011    2.411032    2.385052   \n",
       "AFFX-ThrX-M_at     2.759138    3.055240    2.988927    2.825169    2.891791   \n",
       "AFFX-TrpnX-3_at    1.967994    2.045717    1.961351    2.030025    1.918269   \n",
       "AFFX-TrpnX-5_at    2.791133    3.237114    2.943271    2.863169    2.741495   \n",
       "AFFX-TrpnX-M_at    3.099678    3.405567    3.172511    3.002204    3.038403   \n",
       "\n",
       "                 GSM2612287  \n",
       "ID                           \n",
       "1007_s_at          5.772250  \n",
       "1053_at            7.702891  \n",
       "117_at            10.838637  \n",
       "121_at             6.016455  \n",
       "1255_g_at          2.475537  \n",
       "...                     ...  \n",
       "AFFX-ThrX-5_at     2.573699  \n",
       "AFFX-ThrX-M_at     3.178614  \n",
       "AFFX-TrpnX-3_at    2.202680  \n",
       "AFFX-TrpnX-5_at    3.075356  \n",
       "AFFX-TrpnX-M_at    3.359097  \n",
       "\n",
       "[54675 rows x 192 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10552462</th>\n",
       "      <td>AFFX-ThrX-5_at</td>\n",
       "      <td>2.573698544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10552463</th>\n",
       "      <td>AFFX-ThrX-M_at</td>\n",
       "      <td>3.178614051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10552464</th>\n",
       "      <td>AFFX-TrpnX-3_at</td>\n",
       "      <td>2.202680092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10552465</th>\n",
       "      <td>AFFX-TrpnX-5_at</td>\n",
       "      <td>3.075355563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10552466</th>\n",
       "      <td>AFFX-TrpnX-M_at</td>\n",
       "      <td>3.359096623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10552467 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ID       GB_ACC SPOT_ID Species Scientific Name  \\\n",
       "0               1007_s_at       U48705     NaN            Homo sapiens   \n",
       "1                 1053_at       M87338     NaN            Homo sapiens   \n",
       "2                  117_at       X51757     NaN            Homo sapiens   \n",
       "3                  121_at       X69699     NaN            Homo sapiens   \n",
       "4               1255_g_at       L36861     NaN            Homo sapiens   \n",
       "...                   ...          ...     ...                     ...   \n",
       "10552462   AFFX-ThrX-5_at  2.573698544     NaN                     NaN   \n",
       "10552463   AFFX-ThrX-M_at  3.178614051     NaN                     NaN   \n",
       "10552464  AFFX-TrpnX-3_at  2.202680092     NaN                     NaN   \n",
       "10552465  AFFX-TrpnX-5_at  3.075355563     NaN                     NaN   \n",
       "10552466  AFFX-TrpnX-M_at  3.359096623     NaN                     NaN   \n",
       "\n",
       "         Annotation Date      Sequence Type                  Sequence Source  \\\n",
       "0            Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1            Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "2            Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3            Oct 6, 2014  Exemplar sequence                          GenBank   \n",
       "4            Oct 6, 2014  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "...                  ...                ...                              ...   \n",
       "10552462             NaN                NaN                              NaN   \n",
       "10552463             NaN                NaN                              NaN   \n",
       "10552464             NaN                NaN                              NaN   \n",
       "10552465             NaN                NaN                              NaN   \n",
       "10552466             NaN                NaN                              NaN   \n",
       "\n",
       "                                         Target Description  \\\n",
       "0         U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...   \n",
       "1         M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...   \n",
       "2         X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...   \n",
       "3         X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...   \n",
       "4         L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...   \n",
       "...                                                     ...   \n",
       "10552462                                                NaN   \n",
       "10552463                                                NaN   \n",
       "10552464                                                NaN   \n",
       "10552465                                                NaN   \n",
       "10552466                                                NaN   \n",
       "\n",
       "         Representative Public ID  \\\n",
       "0                          U48705   \n",
       "1                          M87338   \n",
       "2                          X51757   \n",
       "3                          X69699   \n",
       "4                          L36861   \n",
       "...                           ...   \n",
       "10552462                      NaN   \n",
       "10552463                      NaN   \n",
       "10552464                      NaN   \n",
       "10552465                      NaN   \n",
       "10552466                      NaN   \n",
       "\n",
       "                                                 Gene Title       Gene Symbol  \\\n",
       "0         discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1               replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2                      heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                              paired box 8              PAX8   \n",
       "4                   guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "...                                                     ...               ...   \n",
       "10552462                                                NaN               NaN   \n",
       "10552463                                                NaN               NaN   \n",
       "10552464                                                NaN               NaN   \n",
       "10552465                                                NaN               NaN   \n",
       "10552466                                                NaN               NaN   \n",
       "\n",
       "             ENTREZ_GENE_ID  \\\n",
       "0         780 /// 100616237   \n",
       "1                      5982   \n",
       "2                      3310   \n",
       "3                      7849   \n",
       "4                      2978   \n",
       "...                     ...   \n",
       "10552462                NaN   \n",
       "10552463                NaN   \n",
       "10552464                NaN   \n",
       "10552465                NaN   \n",
       "10552466                NaN   \n",
       "\n",
       "                                       RefSeq Transcript ID  \\\n",
       "0         NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1         NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2                                                 NM_002155   \n",
       "3         NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4                                NM_000409 /// XM_006715073   \n",
       "...                                                     ...   \n",
       "10552462                                                NaN   \n",
       "10552463                                                NaN   \n",
       "10552464                                                NaN   \n",
       "10552465                                                NaN   \n",
       "10552466                                                NaN   \n",
       "\n",
       "                           Gene Ontology Biological Process  \\\n",
       "0         0001558 // regulation of cell growth // inferr...   \n",
       "1         0000278 // mitotic cell cycle // traceable aut...   \n",
       "2         0000902 // cell morphogenesis // inferred from...   \n",
       "3         0001655 // urogenital system development // in...   \n",
       "4         0007165 // signal transduction // non-traceabl...   \n",
       "...                                                     ...   \n",
       "10552462                                                NaN   \n",
       "10552463                                                NaN   \n",
       "10552464                                                NaN   \n",
       "10552465                                                NaN   \n",
       "10552466                                                NaN   \n",
       "\n",
       "                           Gene Ontology Cellular Component  \\\n",
       "0         0005576 // extracellular region // inferred fr...   \n",
       "1         0005634 // nucleus // inferred from electronic...   \n",
       "2         0005737 // cytoplasm // inferred from direct a...   \n",
       "3         0005634 // nucleus // inferred from direct ass...   \n",
       "4         0001750 // photoreceptor outer segment // infe...   \n",
       "...                                                     ...   \n",
       "10552462                                                NaN   \n",
       "10552463                                                NaN   \n",
       "10552464                                                NaN   \n",
       "10552465                                                NaN   \n",
       "10552466                                                NaN   \n",
       "\n",
       "                           Gene Ontology Molecular Function  \n",
       "0         0000166 // nucleotide binding // inferred from...  \n",
       "1         0000166 // nucleotide binding // inferred from...  \n",
       "2         0000166 // nucleotide binding // inferred from...  \n",
       "3         0000979 // RNA polymerase II core promoter seq...  \n",
       "4         0005509 // calcium ion binding // inferred fro...  \n",
       "...                                                     ...  \n",
       "10552462                                                NaN  \n",
       "10552463                                                NaN  \n",
       "10552464                                                NaN  \n",
       "10552465                                                NaN  \n",
       "10552466                                                NaN  \n",
       "\n",
       "[10552467 rows x 16 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2612096</th>\n",
       "      <th>GSM2612097</th>\n",
       "      <th>GSM2612098</th>\n",
       "      <th>GSM2612099</th>\n",
       "      <th>GSM2612100</th>\n",
       "      <th>GSM2612101</th>\n",
       "      <th>GSM2612102</th>\n",
       "      <th>GSM2612103</th>\n",
       "      <th>GSM2612104</th>\n",
       "      <th>GSM2612105</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2612278</th>\n",
       "      <th>GSM2612279</th>\n",
       "      <th>GSM2612280</th>\n",
       "      <th>GSM2612281</th>\n",
       "      <th>GSM2612282</th>\n",
       "      <th>GSM2612283</th>\n",
       "      <th>GSM2612284</th>\n",
       "      <th>GSM2612285</th>\n",
       "      <th>GSM2612286</th>\n",
       "      <th>GSM2612287</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>6.887209</td>\n",
       "      <td>6.181964</td>\n",
       "      <td>6.555487</td>\n",
       "      <td>5.873066</td>\n",
       "      <td>5.745004</td>\n",
       "      <td>6.641166</td>\n",
       "      <td>6.464039</td>\n",
       "      <td>6.133586</td>\n",
       "      <td>6.274664</td>\n",
       "      <td>6.054550</td>\n",
       "      <td>...</td>\n",
       "      <td>7.150558</td>\n",
       "      <td>7.406669</td>\n",
       "      <td>7.529819</td>\n",
       "      <td>6.864007</td>\n",
       "      <td>7.001945</td>\n",
       "      <td>7.050386</td>\n",
       "      <td>7.087407</td>\n",
       "      <td>6.896694</td>\n",
       "      <td>7.360574</td>\n",
       "      <td>7.115658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P1</th>\n",
       "      <td>3.770740</td>\n",
       "      <td>3.849153</td>\n",
       "      <td>3.587279</td>\n",
       "      <td>3.026468</td>\n",
       "      <td>2.917429</td>\n",
       "      <td>3.113438</td>\n",
       "      <td>2.945742</td>\n",
       "      <td>3.397430</td>\n",
       "      <td>2.757350</td>\n",
       "      <td>3.446217</td>\n",
       "      <td>...</td>\n",
       "      <td>4.452625</td>\n",
       "      <td>5.330988</td>\n",
       "      <td>4.461095</td>\n",
       "      <td>3.202160</td>\n",
       "      <td>3.611992</td>\n",
       "      <td>4.006498</td>\n",
       "      <td>4.670841</td>\n",
       "      <td>4.149561</td>\n",
       "      <td>3.908489</td>\n",
       "      <td>3.980791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC6P2</th>\n",
       "      <td>3.770740</td>\n",
       "      <td>3.849153</td>\n",
       "      <td>3.587279</td>\n",
       "      <td>3.026468</td>\n",
       "      <td>2.917429</td>\n",
       "      <td>3.113438</td>\n",
       "      <td>2.945742</td>\n",
       "      <td>3.397430</td>\n",
       "      <td>2.757350</td>\n",
       "      <td>3.446217</td>\n",
       "      <td>...</td>\n",
       "      <td>4.452625</td>\n",
       "      <td>5.330988</td>\n",
       "      <td>4.461095</td>\n",
       "      <td>3.202160</td>\n",
       "      <td>3.611992</td>\n",
       "      <td>4.006498</td>\n",
       "      <td>4.670841</td>\n",
       "      <td>4.149561</td>\n",
       "      <td>3.908489</td>\n",
       "      <td>3.980791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1P2</th>\n",
       "      <td>3.671268</td>\n",
       "      <td>3.251920</td>\n",
       "      <td>2.888037</td>\n",
       "      <td>3.098870</td>\n",
       "      <td>3.274688</td>\n",
       "      <td>3.004873</td>\n",
       "      <td>2.793291</td>\n",
       "      <td>3.240455</td>\n",
       "      <td>3.303735</td>\n",
       "      <td>3.518894</td>\n",
       "      <td>...</td>\n",
       "      <td>3.088957</td>\n",
       "      <td>3.423271</td>\n",
       "      <td>3.088001</td>\n",
       "      <td>3.194958</td>\n",
       "      <td>3.352198</td>\n",
       "      <td>3.436326</td>\n",
       "      <td>3.176444</td>\n",
       "      <td>3.377824</td>\n",
       "      <td>3.152157</td>\n",
       "      <td>3.445710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC078883.4</th>\n",
       "      <td>4.540082</td>\n",
       "      <td>4.221583</td>\n",
       "      <td>4.406263</td>\n",
       "      <td>3.895509</td>\n",
       "      <td>4.212557</td>\n",
       "      <td>3.518138</td>\n",
       "      <td>3.875489</td>\n",
       "      <td>4.169540</td>\n",
       "      <td>4.052477</td>\n",
       "      <td>4.282009</td>\n",
       "      <td>...</td>\n",
       "      <td>4.374490</td>\n",
       "      <td>4.178730</td>\n",
       "      <td>4.138277</td>\n",
       "      <td>4.331086</td>\n",
       "      <td>4.361535</td>\n",
       "      <td>4.509868</td>\n",
       "      <td>4.303939</td>\n",
       "      <td>4.293775</td>\n",
       "      <td>4.281061</td>\n",
       "      <td>4.637451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>abParts</th>\n",
       "      <td>6.840695</td>\n",
       "      <td>6.761038</td>\n",
       "      <td>8.097322</td>\n",
       "      <td>7.367817</td>\n",
       "      <td>5.648297</td>\n",
       "      <td>7.888521</td>\n",
       "      <td>6.200375</td>\n",
       "      <td>6.745986</td>\n",
       "      <td>6.421771</td>\n",
       "      <td>7.483332</td>\n",
       "      <td>...</td>\n",
       "      <td>7.742626</td>\n",
       "      <td>8.184422</td>\n",
       "      <td>8.277556</td>\n",
       "      <td>7.479552</td>\n",
       "      <td>8.518359</td>\n",
       "      <td>7.365643</td>\n",
       "      <td>8.075855</td>\n",
       "      <td>7.948421</td>\n",
       "      <td>8.453753</td>\n",
       "      <td>7.860338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alpha</th>\n",
       "      <td>7.993384</td>\n",
       "      <td>7.859073</td>\n",
       "      <td>7.443326</td>\n",
       "      <td>7.288378</td>\n",
       "      <td>5.391930</td>\n",
       "      <td>7.254210</td>\n",
       "      <td>6.638218</td>\n",
       "      <td>7.073584</td>\n",
       "      <td>7.923679</td>\n",
       "      <td>7.498770</td>\n",
       "      <td>...</td>\n",
       "      <td>6.715298</td>\n",
       "      <td>9.210556</td>\n",
       "      <td>8.251305</td>\n",
       "      <td>6.888582</td>\n",
       "      <td>8.485592</td>\n",
       "      <td>7.241490</td>\n",
       "      <td>8.236890</td>\n",
       "      <td>8.081402</td>\n",
       "      <td>8.295665</td>\n",
       "      <td>8.086276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>av27s1</th>\n",
       "      <td>3.267696</td>\n",
       "      <td>5.357518</td>\n",
       "      <td>4.489498</td>\n",
       "      <td>5.203069</td>\n",
       "      <td>5.114636</td>\n",
       "      <td>4.838910</td>\n",
       "      <td>4.514805</td>\n",
       "      <td>4.865177</td>\n",
       "      <td>4.920292</td>\n",
       "      <td>4.544280</td>\n",
       "      <td>...</td>\n",
       "      <td>4.951748</td>\n",
       "      <td>5.418962</td>\n",
       "      <td>5.442874</td>\n",
       "      <td>5.258984</td>\n",
       "      <td>5.787806</td>\n",
       "      <td>4.993456</td>\n",
       "      <td>4.926423</td>\n",
       "      <td>6.270793</td>\n",
       "      <td>5.380123</td>\n",
       "      <td>4.624921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>4.173511</td>\n",
       "      <td>3.739451</td>\n",
       "      <td>3.689065</td>\n",
       "      <td>3.757212</td>\n",
       "      <td>4.251514</td>\n",
       "      <td>4.259409</td>\n",
       "      <td>3.850559</td>\n",
       "      <td>4.035009</td>\n",
       "      <td>4.156332</td>\n",
       "      <td>3.819965</td>\n",
       "      <td>...</td>\n",
       "      <td>3.731068</td>\n",
       "      <td>3.860831</td>\n",
       "      <td>3.787460</td>\n",
       "      <td>3.719467</td>\n",
       "      <td>3.674982</td>\n",
       "      <td>3.980256</td>\n",
       "      <td>3.852515</td>\n",
       "      <td>3.792280</td>\n",
       "      <td>3.986952</td>\n",
       "      <td>3.899352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mir-223</th>\n",
       "      <td>10.175704</td>\n",
       "      <td>10.748655</td>\n",
       "      <td>10.314610</td>\n",
       "      <td>10.708885</td>\n",
       "      <td>10.165887</td>\n",
       "      <td>10.572054</td>\n",
       "      <td>10.104610</td>\n",
       "      <td>10.650281</td>\n",
       "      <td>10.907954</td>\n",
       "      <td>10.564287</td>\n",
       "      <td>...</td>\n",
       "      <td>10.859423</td>\n",
       "      <td>10.858151</td>\n",
       "      <td>10.138659</td>\n",
       "      <td>10.817622</td>\n",
       "      <td>10.526606</td>\n",
       "      <td>10.961939</td>\n",
       "      <td>10.982302</td>\n",
       "      <td>10.628225</td>\n",
       "      <td>10.764013</td>\n",
       "      <td>10.784081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25851 rows Ã— 192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GSM2612096  GSM2612097  GSM2612098  GSM2612099  GSM2612100  \\\n",
       "Gene                                                                        \n",
       " ABCB4           6.887209    6.181964    6.555487    5.873066    5.745004   \n",
       " ABCC6P1         3.770740    3.849153    3.587279    3.026468    2.917429   \n",
       " ABCC6P2         3.770740    3.849153    3.587279    3.026468    2.917429   \n",
       " ABCD1P2         3.671268    3.251920    2.888037    3.098870    3.274688   \n",
       " AC078883.4      4.540082    4.221583    4.406263    3.895509    4.212557   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts          6.840695    6.761038    8.097322    7.367817    5.648297   \n",
       "alpha            7.993384    7.859073    7.443326    7.288378    5.391930   \n",
       "av27s1           3.267696    5.357518    4.489498    5.203069    5.114636   \n",
       "hsa-let-7a-3     4.173511    3.739451    3.689065    3.757212    4.251514   \n",
       "mir-223         10.175704   10.748655   10.314610   10.708885   10.165887   \n",
       "\n",
       "               GSM2612101  GSM2612102  GSM2612103  GSM2612104  GSM2612105  \\\n",
       "Gene                                                                        \n",
       " ABCB4           6.641166    6.464039    6.133586    6.274664    6.054550   \n",
       " ABCC6P1         3.113438    2.945742    3.397430    2.757350    3.446217   \n",
       " ABCC6P2         3.113438    2.945742    3.397430    2.757350    3.446217   \n",
       " ABCD1P2         3.004873    2.793291    3.240455    3.303735    3.518894   \n",
       " AC078883.4      3.518138    3.875489    4.169540    4.052477    4.282009   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts          7.888521    6.200375    6.745986    6.421771    7.483332   \n",
       "alpha            7.254210    6.638218    7.073584    7.923679    7.498770   \n",
       "av27s1           4.838910    4.514805    4.865177    4.920292    4.544280   \n",
       "hsa-let-7a-3     4.259409    3.850559    4.035009    4.156332    3.819965   \n",
       "mir-223         10.572054   10.104610   10.650281   10.907954   10.564287   \n",
       "\n",
       "               ...  GSM2612278  GSM2612279  GSM2612280  GSM2612281  \\\n",
       "Gene           ...                                                   \n",
       " ABCB4         ...    7.150558    7.406669    7.529819    6.864007   \n",
       " ABCC6P1       ...    4.452625    5.330988    4.461095    3.202160   \n",
       " ABCC6P2       ...    4.452625    5.330988    4.461095    3.202160   \n",
       " ABCD1P2       ...    3.088957    3.423271    3.088001    3.194958   \n",
       " AC078883.4    ...    4.374490    4.178730    4.138277    4.331086   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "abParts        ...    7.742626    8.184422    8.277556    7.479552   \n",
       "alpha          ...    6.715298    9.210556    8.251305    6.888582   \n",
       "av27s1         ...    4.951748    5.418962    5.442874    5.258984   \n",
       "hsa-let-7a-3   ...    3.731068    3.860831    3.787460    3.719467   \n",
       "mir-223        ...   10.859423   10.858151   10.138659   10.817622   \n",
       "\n",
       "               GSM2612282  GSM2612283  GSM2612284  GSM2612285  GSM2612286  \\\n",
       "Gene                                                                        \n",
       " ABCB4           7.001945    7.050386    7.087407    6.896694    7.360574   \n",
       " ABCC6P1         3.611992    4.006498    4.670841    4.149561    3.908489   \n",
       " ABCC6P2         3.611992    4.006498    4.670841    4.149561    3.908489   \n",
       " ABCD1P2         3.352198    3.436326    3.176444    3.377824    3.152157   \n",
       " AC078883.4      4.361535    4.509868    4.303939    4.293775    4.281061   \n",
       "...                   ...         ...         ...         ...         ...   \n",
       "abParts          8.518359    7.365643    8.075855    7.948421    8.453753   \n",
       "alpha            8.485592    7.241490    8.236890    8.081402    8.295665   \n",
       "av27s1           5.787806    4.993456    4.926423    6.270793    5.380123   \n",
       "hsa-let-7a-3     3.674982    3.980256    3.852515    3.792280    3.986952   \n",
       "mir-223         10.526606   10.961939   10.982302   10.628225   10.764013   \n",
       "\n",
       "               GSM2612287  \n",
       "Gene                       \n",
       " ABCB4           7.115658  \n",
       " ABCC6P1         3.980791  \n",
       " ABCC6P2         3.980791  \n",
       " ABCD1P2         3.445710  \n",
       " AC078883.4      4.637451  \n",
       "...                   ...  \n",
       "abParts          7.860338  \n",
       "alpha            8.086276  \n",
       "av27s1           4.624921  \n",
       "hsa-let-7a-3     3.899352  \n",
       "mir-223         10.784081  \n",
       "\n",
       "[25851 rows x 192 columns]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2612096</th>\n",
       "      <th>GSM2612097</th>\n",
       "      <th>GSM2612098</th>\n",
       "      <th>GSM2612099</th>\n",
       "      <th>GSM2612100</th>\n",
       "      <th>GSM2612101</th>\n",
       "      <th>GSM2612102</th>\n",
       "      <th>GSM2612103</th>\n",
       "      <th>GSM2612104</th>\n",
       "      <th>GSM2612105</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2612278</th>\n",
       "      <th>GSM2612279</th>\n",
       "      <th>GSM2612280</th>\n",
       "      <th>GSM2612281</th>\n",
       "      <th>GSM2612282</th>\n",
       "      <th>GSM2612283</th>\n",
       "      <th>GSM2612284</th>\n",
       "      <th>GSM2612285</th>\n",
       "      <th>GSM2612286</th>\n",
       "      <th>GSM2612287</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>3.678218</td>\n",
       "      <td>3.286122</td>\n",
       "      <td>2.978241</td>\n",
       "      <td>3.416928</td>\n",
       "      <td>3.915461</td>\n",
       "      <td>3.622674</td>\n",
       "      <td>3.348873</td>\n",
       "      <td>3.407293</td>\n",
       "      <td>3.475643</td>\n",
       "      <td>3.588239</td>\n",
       "      <td>...</td>\n",
       "      <td>3.845511</td>\n",
       "      <td>3.589142</td>\n",
       "      <td>3.502773</td>\n",
       "      <td>3.322457</td>\n",
       "      <td>3.618603</td>\n",
       "      <td>3.791884</td>\n",
       "      <td>3.453535</td>\n",
       "      <td>3.363853</td>\n",
       "      <td>3.454187</td>\n",
       "      <td>3.765669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>6.443778</td>\n",
       "      <td>6.439185</td>\n",
       "      <td>6.091514</td>\n",
       "      <td>6.160160</td>\n",
       "      <td>6.262001</td>\n",
       "      <td>6.650213</td>\n",
       "      <td>6.046788</td>\n",
       "      <td>6.930769</td>\n",
       "      <td>6.418877</td>\n",
       "      <td>6.679607</td>\n",
       "      <td>...</td>\n",
       "      <td>6.482534</td>\n",
       "      <td>6.767941</td>\n",
       "      <td>6.677439</td>\n",
       "      <td>6.602510</td>\n",
       "      <td>6.787397</td>\n",
       "      <td>7.079915</td>\n",
       "      <td>6.603517</td>\n",
       "      <td>6.683437</td>\n",
       "      <td>6.562028</td>\n",
       "      <td>7.033624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>3.282722</td>\n",
       "      <td>2.779510</td>\n",
       "      <td>2.701834</td>\n",
       "      <td>3.136804</td>\n",
       "      <td>2.722878</td>\n",
       "      <td>2.927107</td>\n",
       "      <td>2.806098</td>\n",
       "      <td>2.691950</td>\n",
       "      <td>2.971447</td>\n",
       "      <td>2.958255</td>\n",
       "      <td>...</td>\n",
       "      <td>2.643486</td>\n",
       "      <td>3.209170</td>\n",
       "      <td>2.591007</td>\n",
       "      <td>2.900855</td>\n",
       "      <td>2.649972</td>\n",
       "      <td>3.033433</td>\n",
       "      <td>2.871597</td>\n",
       "      <td>2.804416</td>\n",
       "      <td>2.773500</td>\n",
       "      <td>3.082072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>4.952460</td>\n",
       "      <td>4.575449</td>\n",
       "      <td>4.801654</td>\n",
       "      <td>4.814918</td>\n",
       "      <td>4.836391</td>\n",
       "      <td>4.607112</td>\n",
       "      <td>4.566239</td>\n",
       "      <td>4.528694</td>\n",
       "      <td>4.940149</td>\n",
       "      <td>4.757857</td>\n",
       "      <td>...</td>\n",
       "      <td>4.389326</td>\n",
       "      <td>4.522569</td>\n",
       "      <td>4.750834</td>\n",
       "      <td>4.543339</td>\n",
       "      <td>4.872960</td>\n",
       "      <td>4.662414</td>\n",
       "      <td>4.716106</td>\n",
       "      <td>4.468409</td>\n",
       "      <td>4.588417</td>\n",
       "      <td>4.876001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>8.313736</td>\n",
       "      <td>7.442635</td>\n",
       "      <td>8.522661</td>\n",
       "      <td>8.584944</td>\n",
       "      <td>8.582563</td>\n",
       "      <td>8.080346</td>\n",
       "      <td>8.787317</td>\n",
       "      <td>8.089630</td>\n",
       "      <td>7.931064</td>\n",
       "      <td>7.702357</td>\n",
       "      <td>...</td>\n",
       "      <td>7.133639</td>\n",
       "      <td>7.295100</td>\n",
       "      <td>8.266381</td>\n",
       "      <td>6.530439</td>\n",
       "      <td>7.978626</td>\n",
       "      <td>7.685621</td>\n",
       "      <td>7.938541</td>\n",
       "      <td>7.261100</td>\n",
       "      <td>8.017463</td>\n",
       "      <td>8.066153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>4.472377</td>\n",
       "      <td>4.687464</td>\n",
       "      <td>4.928949</td>\n",
       "      <td>4.379016</td>\n",
       "      <td>4.538363</td>\n",
       "      <td>4.160522</td>\n",
       "      <td>4.396658</td>\n",
       "      <td>4.667440</td>\n",
       "      <td>4.280527</td>\n",
       "      <td>3.729029</td>\n",
       "      <td>...</td>\n",
       "      <td>3.390104</td>\n",
       "      <td>3.393494</td>\n",
       "      <td>4.065496</td>\n",
       "      <td>3.282047</td>\n",
       "      <td>3.713914</td>\n",
       "      <td>3.856949</td>\n",
       "      <td>3.856006</td>\n",
       "      <td>3.690631</td>\n",
       "      <td>3.736156</td>\n",
       "      <td>4.090386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>7.702434</td>\n",
       "      <td>8.273578</td>\n",
       "      <td>8.332399</td>\n",
       "      <td>8.055039</td>\n",
       "      <td>8.177309</td>\n",
       "      <td>8.418922</td>\n",
       "      <td>8.559867</td>\n",
       "      <td>7.807338</td>\n",
       "      <td>8.284237</td>\n",
       "      <td>7.906655</td>\n",
       "      <td>...</td>\n",
       "      <td>8.116576</td>\n",
       "      <td>8.004243</td>\n",
       "      <td>7.810824</td>\n",
       "      <td>7.475349</td>\n",
       "      <td>7.764784</td>\n",
       "      <td>7.792606</td>\n",
       "      <td>7.508345</td>\n",
       "      <td>7.978566</td>\n",
       "      <td>7.914982</td>\n",
       "      <td>7.472140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>9.922776</td>\n",
       "      <td>9.761252</td>\n",
       "      <td>8.414962</td>\n",
       "      <td>9.092369</td>\n",
       "      <td>8.967648</td>\n",
       "      <td>9.341418</td>\n",
       "      <td>8.551150</td>\n",
       "      <td>9.326148</td>\n",
       "      <td>9.529829</td>\n",
       "      <td>9.574537</td>\n",
       "      <td>...</td>\n",
       "      <td>10.017479</td>\n",
       "      <td>10.863439</td>\n",
       "      <td>9.639776</td>\n",
       "      <td>10.481763</td>\n",
       "      <td>10.356037</td>\n",
       "      <td>10.074941</td>\n",
       "      <td>9.985552</td>\n",
       "      <td>10.279970</td>\n",
       "      <td>9.808788</td>\n",
       "      <td>10.101097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>6.453670</td>\n",
       "      <td>6.335933</td>\n",
       "      <td>6.446376</td>\n",
       "      <td>6.527935</td>\n",
       "      <td>6.428279</td>\n",
       "      <td>6.555135</td>\n",
       "      <td>6.586912</td>\n",
       "      <td>6.558270</td>\n",
       "      <td>6.884512</td>\n",
       "      <td>6.341421</td>\n",
       "      <td>...</td>\n",
       "      <td>6.501560</td>\n",
       "      <td>6.853390</td>\n",
       "      <td>6.646841</td>\n",
       "      <td>6.643384</td>\n",
       "      <td>6.619929</td>\n",
       "      <td>6.850845</td>\n",
       "      <td>6.910645</td>\n",
       "      <td>6.938110</td>\n",
       "      <td>6.824424</td>\n",
       "      <td>6.589108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>4.693944</td>\n",
       "      <td>5.581965</td>\n",
       "      <td>6.012590</td>\n",
       "      <td>5.716714</td>\n",
       "      <td>5.297856</td>\n",
       "      <td>5.759125</td>\n",
       "      <td>6.041967</td>\n",
       "      <td>5.032671</td>\n",
       "      <td>5.168541</td>\n",
       "      <td>4.633298</td>\n",
       "      <td>...</td>\n",
       "      <td>5.925068</td>\n",
       "      <td>5.344623</td>\n",
       "      <td>6.382690</td>\n",
       "      <td>5.762678</td>\n",
       "      <td>5.864292</td>\n",
       "      <td>5.711944</td>\n",
       "      <td>5.948168</td>\n",
       "      <td>5.699622</td>\n",
       "      <td>6.270375</td>\n",
       "      <td>5.675278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows Ã— 192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM2612096  GSM2612097  GSM2612098  GSM2612099  GSM2612100  \\\n",
       "A1BG        3.678218    3.286122    2.978241    3.416928    3.915461   \n",
       "A1BG-AS1    6.443778    6.439185    6.091514    6.160160    6.262001   \n",
       "A1CF        3.282722    2.779510    2.701834    3.136804    2.722878   \n",
       "A2M         4.952460    4.575449    4.801654    4.814918    4.836391   \n",
       "A2M-AS1     8.313736    7.442635    8.522661    8.584944    8.582563   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A      4.472377    4.687464    4.928949    4.379016    4.538363   \n",
       "ZYG11B      7.702434    8.273578    8.332399    8.055039    8.177309   \n",
       "ZYX         9.922776    9.761252    8.414962    9.092369    8.967648   \n",
       "ZZEF1       6.453670    6.335933    6.446376    6.527935    6.428279   \n",
       "ZZZ3        4.693944    5.581965    6.012590    5.716714    5.297856   \n",
       "\n",
       "          GSM2612101  GSM2612102  GSM2612103  GSM2612104  GSM2612105  ...  \\\n",
       "A1BG        3.622674    3.348873    3.407293    3.475643    3.588239  ...   \n",
       "A1BG-AS1    6.650213    6.046788    6.930769    6.418877    6.679607  ...   \n",
       "A1CF        2.927107    2.806098    2.691950    2.971447    2.958255  ...   \n",
       "A2M         4.607112    4.566239    4.528694    4.940149    4.757857  ...   \n",
       "A2M-AS1     8.080346    8.787317    8.089630    7.931064    7.702357  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A      4.160522    4.396658    4.667440    4.280527    3.729029  ...   \n",
       "ZYG11B      8.418922    8.559867    7.807338    8.284237    7.906655  ...   \n",
       "ZYX         9.341418    8.551150    9.326148    9.529829    9.574537  ...   \n",
       "ZZEF1       6.555135    6.586912    6.558270    6.884512    6.341421  ...   \n",
       "ZZZ3        5.759125    6.041967    5.032671    5.168541    4.633298  ...   \n",
       "\n",
       "          GSM2612278  GSM2612279  GSM2612280  GSM2612281  GSM2612282  \\\n",
       "A1BG        3.845511    3.589142    3.502773    3.322457    3.618603   \n",
       "A1BG-AS1    6.482534    6.767941    6.677439    6.602510    6.787397   \n",
       "A1CF        2.643486    3.209170    2.591007    2.900855    2.649972   \n",
       "A2M         4.389326    4.522569    4.750834    4.543339    4.872960   \n",
       "A2M-AS1     7.133639    7.295100    8.266381    6.530439    7.978626   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A      3.390104    3.393494    4.065496    3.282047    3.713914   \n",
       "ZYG11B      8.116576    8.004243    7.810824    7.475349    7.764784   \n",
       "ZYX        10.017479   10.863439    9.639776   10.481763   10.356037   \n",
       "ZZEF1       6.501560    6.853390    6.646841    6.643384    6.619929   \n",
       "ZZZ3        5.925068    5.344623    6.382690    5.762678    5.864292   \n",
       "\n",
       "          GSM2612283  GSM2612284  GSM2612285  GSM2612286  GSM2612287  \n",
       "A1BG        3.791884    3.453535    3.363853    3.454187    3.765669  \n",
       "A1BG-AS1    7.079915    6.603517    6.683437    6.562028    7.033624  \n",
       "A1CF        3.033433    2.871597    2.804416    2.773500    3.082072  \n",
       "A2M         4.662414    4.716106    4.468409    4.588417    4.876001  \n",
       "A2M-AS1     7.685621    7.938541    7.261100    8.017463    8.066153  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "ZYG11A      3.856949    3.856006    3.690631    3.736156    4.090386  \n",
       "ZYG11B      7.792606    7.508345    7.978566    7.914982    7.472140  \n",
       "ZYX        10.074941    9.985552   10.279970    9.808788   10.101097  \n",
       "ZZEF1       6.850845    6.910645    6.938110    6.824424    6.589108  \n",
       "ZZZ3        5.711944    5.948168    5.699622    6.270375    5.675278  \n",
       "\n",
       "[17857 rows x 192 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2612096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>35.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.678218</td>\n",
       "      <td>6.443778</td>\n",
       "      <td>3.282722</td>\n",
       "      <td>4.952460</td>\n",
       "      <td>8.313736</td>\n",
       "      <td>2.896964</td>\n",
       "      <td>5.992351</td>\n",
       "      <td>...</td>\n",
       "      <td>4.365722</td>\n",
       "      <td>6.239195</td>\n",
       "      <td>5.691087</td>\n",
       "      <td>3.869996</td>\n",
       "      <td>6.513365</td>\n",
       "      <td>4.472377</td>\n",
       "      <td>7.702434</td>\n",
       "      <td>9.922776</td>\n",
       "      <td>6.453670</td>\n",
       "      <td>4.693944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>36.9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.286122</td>\n",
       "      <td>6.439185</td>\n",
       "      <td>2.779510</td>\n",
       "      <td>4.575449</td>\n",
       "      <td>7.442635</td>\n",
       "      <td>2.811418</td>\n",
       "      <td>6.043174</td>\n",
       "      <td>...</td>\n",
       "      <td>4.564470</td>\n",
       "      <td>5.821275</td>\n",
       "      <td>5.678074</td>\n",
       "      <td>4.446895</td>\n",
       "      <td>6.941676</td>\n",
       "      <td>4.687464</td>\n",
       "      <td>8.273578</td>\n",
       "      <td>9.761252</td>\n",
       "      <td>6.335933</td>\n",
       "      <td>5.581965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612098</th>\n",
       "      <td>0.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.978241</td>\n",
       "      <td>6.091514</td>\n",
       "      <td>2.701834</td>\n",
       "      <td>4.801654</td>\n",
       "      <td>8.522661</td>\n",
       "      <td>2.564243</td>\n",
       "      <td>6.885544</td>\n",
       "      <td>...</td>\n",
       "      <td>5.097691</td>\n",
       "      <td>7.184773</td>\n",
       "      <td>6.748653</td>\n",
       "      <td>5.092760</td>\n",
       "      <td>7.343632</td>\n",
       "      <td>4.928949</td>\n",
       "      <td>8.332399</td>\n",
       "      <td>8.414962</td>\n",
       "      <td>6.446376</td>\n",
       "      <td>6.012590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612099</th>\n",
       "      <td>0.0</td>\n",
       "      <td>35.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.416928</td>\n",
       "      <td>6.160160</td>\n",
       "      <td>3.136804</td>\n",
       "      <td>4.814918</td>\n",
       "      <td>8.584944</td>\n",
       "      <td>2.754305</td>\n",
       "      <td>6.822452</td>\n",
       "      <td>...</td>\n",
       "      <td>5.084126</td>\n",
       "      <td>6.549417</td>\n",
       "      <td>5.525266</td>\n",
       "      <td>4.292599</td>\n",
       "      <td>7.140479</td>\n",
       "      <td>4.379016</td>\n",
       "      <td>8.055039</td>\n",
       "      <td>9.092369</td>\n",
       "      <td>6.527935</td>\n",
       "      <td>5.716714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612100</th>\n",
       "      <td>0.0</td>\n",
       "      <td>53.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.915461</td>\n",
       "      <td>6.262001</td>\n",
       "      <td>2.722878</td>\n",
       "      <td>4.836391</td>\n",
       "      <td>8.582563</td>\n",
       "      <td>3.218403</td>\n",
       "      <td>7.374226</td>\n",
       "      <td>...</td>\n",
       "      <td>4.983173</td>\n",
       "      <td>6.058277</td>\n",
       "      <td>6.216823</td>\n",
       "      <td>4.518410</td>\n",
       "      <td>6.689277</td>\n",
       "      <td>4.538363</td>\n",
       "      <td>8.177309</td>\n",
       "      <td>8.967648</td>\n",
       "      <td>6.428279</td>\n",
       "      <td>5.297856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612283</th>\n",
       "      <td>0.0</td>\n",
       "      <td>62.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.791884</td>\n",
       "      <td>7.079915</td>\n",
       "      <td>3.033433</td>\n",
       "      <td>4.662414</td>\n",
       "      <td>7.685621</td>\n",
       "      <td>2.400548</td>\n",
       "      <td>4.939600</td>\n",
       "      <td>...</td>\n",
       "      <td>4.744505</td>\n",
       "      <td>5.960281</td>\n",
       "      <td>5.231191</td>\n",
       "      <td>4.844246</td>\n",
       "      <td>7.182306</td>\n",
       "      <td>3.856949</td>\n",
       "      <td>7.792606</td>\n",
       "      <td>10.074941</td>\n",
       "      <td>6.850845</td>\n",
       "      <td>5.711944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612284</th>\n",
       "      <td>0.0</td>\n",
       "      <td>73.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.453535</td>\n",
       "      <td>6.603517</td>\n",
       "      <td>2.871597</td>\n",
       "      <td>4.716106</td>\n",
       "      <td>7.938541</td>\n",
       "      <td>2.448653</td>\n",
       "      <td>5.530573</td>\n",
       "      <td>...</td>\n",
       "      <td>4.965975</td>\n",
       "      <td>6.443442</td>\n",
       "      <td>5.754605</td>\n",
       "      <td>4.754218</td>\n",
       "      <td>7.200340</td>\n",
       "      <td>3.856006</td>\n",
       "      <td>7.508345</td>\n",
       "      <td>9.985552</td>\n",
       "      <td>6.910645</td>\n",
       "      <td>5.948168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612285</th>\n",
       "      <td>1.0</td>\n",
       "      <td>61.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.363853</td>\n",
       "      <td>6.683437</td>\n",
       "      <td>2.804416</td>\n",
       "      <td>4.468409</td>\n",
       "      <td>7.261100</td>\n",
       "      <td>2.562233</td>\n",
       "      <td>5.239228</td>\n",
       "      <td>...</td>\n",
       "      <td>4.895647</td>\n",
       "      <td>6.270400</td>\n",
       "      <td>5.716105</td>\n",
       "      <td>4.821335</td>\n",
       "      <td>7.223925</td>\n",
       "      <td>3.690631</td>\n",
       "      <td>7.978566</td>\n",
       "      <td>10.279970</td>\n",
       "      <td>6.938110</td>\n",
       "      <td>5.699622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612286</th>\n",
       "      <td>0.0</td>\n",
       "      <td>31.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.454187</td>\n",
       "      <td>6.562028</td>\n",
       "      <td>2.773500</td>\n",
       "      <td>4.588417</td>\n",
       "      <td>8.017463</td>\n",
       "      <td>2.184304</td>\n",
       "      <td>5.584945</td>\n",
       "      <td>...</td>\n",
       "      <td>5.111829</td>\n",
       "      <td>7.238046</td>\n",
       "      <td>5.936044</td>\n",
       "      <td>5.046152</td>\n",
       "      <td>7.297916</td>\n",
       "      <td>3.736156</td>\n",
       "      <td>7.914982</td>\n",
       "      <td>9.808788</td>\n",
       "      <td>6.824424</td>\n",
       "      <td>6.270375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2612287</th>\n",
       "      <td>0.0</td>\n",
       "      <td>61.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.765669</td>\n",
       "      <td>7.033624</td>\n",
       "      <td>3.082072</td>\n",
       "      <td>4.876001</td>\n",
       "      <td>8.066153</td>\n",
       "      <td>2.508471</td>\n",
       "      <td>4.949683</td>\n",
       "      <td>...</td>\n",
       "      <td>4.812881</td>\n",
       "      <td>5.576477</td>\n",
       "      <td>5.760265</td>\n",
       "      <td>4.852928</td>\n",
       "      <td>7.002750</td>\n",
       "      <td>4.090386</td>\n",
       "      <td>7.472140</td>\n",
       "      <td>10.101097</td>\n",
       "      <td>6.589108</td>\n",
       "      <td>5.675278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows Ã— 17860 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Anxiety disorder   Age  Gender      A1BG  A1BG-AS1      A1CF  \\\n",
       "GSM2612096               0.0  35.8     0.0  3.678218  6.443778  3.282722   \n",
       "GSM2612097               0.0  36.9     1.0  3.286122  6.439185  2.779510   \n",
       "GSM2612098               0.0  62.0     1.0  2.978241  6.091514  2.701834   \n",
       "GSM2612099               0.0  35.7     1.0  3.416928  6.160160  3.136804   \n",
       "GSM2612100               0.0  53.4     0.0  3.915461  6.262001  2.722878   \n",
       "...                      ...   ...     ...       ...       ...       ...   \n",
       "GSM2612283               0.0  62.3     1.0  3.791884  7.079915  3.033433   \n",
       "GSM2612284               0.0  73.1     1.0  3.453535  6.603517  2.871597   \n",
       "GSM2612285               1.0  61.3     1.0  3.363853  6.683437  2.804416   \n",
       "GSM2612286               0.0  31.2     1.0  3.454187  6.562028  2.773500   \n",
       "GSM2612287               0.0  61.2     1.0  3.765669  7.033624  3.082072   \n",
       "\n",
       "                 A2M   A2M-AS1     A2ML1     A2MP1  ...    ZWILCH     ZWINT  \\\n",
       "GSM2612096  4.952460  8.313736  2.896964  5.992351  ...  4.365722  6.239195   \n",
       "GSM2612097  4.575449  7.442635  2.811418  6.043174  ...  4.564470  5.821275   \n",
       "GSM2612098  4.801654  8.522661  2.564243  6.885544  ...  5.097691  7.184773   \n",
       "GSM2612099  4.814918  8.584944  2.754305  6.822452  ...  5.084126  6.549417   \n",
       "GSM2612100  4.836391  8.582563  3.218403  7.374226  ...  4.983173  6.058277   \n",
       "...              ...       ...       ...       ...  ...       ...       ...   \n",
       "GSM2612283  4.662414  7.685621  2.400548  4.939600  ...  4.744505  5.960281   \n",
       "GSM2612284  4.716106  7.938541  2.448653  5.530573  ...  4.965975  6.443442   \n",
       "GSM2612285  4.468409  7.261100  2.562233  5.239228  ...  4.895647  6.270400   \n",
       "GSM2612286  4.588417  8.017463  2.184304  5.584945  ...  5.111829  7.238046   \n",
       "GSM2612287  4.876001  8.066153  2.508471  4.949683  ...  4.812881  5.576477   \n",
       "\n",
       "                ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B        ZYX  \\\n",
       "GSM2612096  5.691087  3.869996  6.513365  4.472377  7.702434   9.922776   \n",
       "GSM2612097  5.678074  4.446895  6.941676  4.687464  8.273578   9.761252   \n",
       "GSM2612098  6.748653  5.092760  7.343632  4.928949  8.332399   8.414962   \n",
       "GSM2612099  5.525266  4.292599  7.140479  4.379016  8.055039   9.092369   \n",
       "GSM2612100  6.216823  4.518410  6.689277  4.538363  8.177309   8.967648   \n",
       "...              ...       ...       ...       ...       ...        ...   \n",
       "GSM2612283  5.231191  4.844246  7.182306  3.856949  7.792606  10.074941   \n",
       "GSM2612284  5.754605  4.754218  7.200340  3.856006  7.508345   9.985552   \n",
       "GSM2612285  5.716105  4.821335  7.223925  3.690631  7.978566  10.279970   \n",
       "GSM2612286  5.936044  5.046152  7.297916  3.736156  7.914982   9.808788   \n",
       "GSM2612287  5.760265  4.852928  7.002750  4.090386  7.472140  10.101097   \n",
       "\n",
       "               ZZEF1      ZZZ3  \n",
       "GSM2612096  6.453670  4.693944  \n",
       "GSM2612097  6.335933  5.581965  \n",
       "GSM2612098  6.446376  6.012590  \n",
       "GSM2612099  6.527935  5.716714  \n",
       "GSM2612100  6.428279  5.297856  \n",
       "...              ...       ...  \n",
       "GSM2612283  6.850845  5.711944  \n",
       "GSM2612284  6.910645  5.948168  \n",
       "GSM2612285  6.938110  5.699622  \n",
       "GSM2612286  6.824424  6.270375  \n",
       "GSM2612287  6.589108  5.675278  \n",
       "\n",
       "[192 rows x 17860 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 192 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1.0' with 64 occurrences. This represents 33.33% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 43.875\n",
      "  50% (Median): 52.6\n",
      "  75%: 62.05\n",
      "Min: 31.0\n",
      "Max: 73.1\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 48 occurrences. This represents 25.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Blood gene expression profiles associated with symptoms of generalized anxiety disorder\"\n",
      "!Series_summary\t\"Prospective epidemiological studies found that generalized anxiety disorder (GAD) can impair immune function and increase risk for cardiovascular disease or events. Mechanisms underlying the physiological reververations of anxiety, however, are still elusive. Hence, we aimed to investigate molecular processes mediating effects of anxiety on physical health using blood gene expression profiles of 546 community participants. Of these, 179 met the status of controls and 157 cases of anxiety.\"\n",
      "!Series_overall_design\t\"We examined genome-wide differential gene expression in anxiety, as well as associations between nine major modules of co-regulated transcripts in blood gene expression and anxiety. There were a total of 546 subjects.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1510561</th>\n",
       "      <th>GSM1510562</th>\n",
       "      <th>GSM1510563</th>\n",
       "      <th>GSM1510564</th>\n",
       "      <th>GSM1510565</th>\n",
       "      <th>GSM1510566</th>\n",
       "      <th>GSM1510567</th>\n",
       "      <th>GSM1510568</th>\n",
       "      <th>GSM1510569</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1511097</th>\n",
       "      <th>GSM1511098</th>\n",
       "      <th>GSM1511099</th>\n",
       "      <th>GSM1511100</th>\n",
       "      <th>GSM1511101</th>\n",
       "      <th>GSM1511102</th>\n",
       "      <th>GSM1511103</th>\n",
       "      <th>GSM1511104</th>\n",
       "      <th>GSM1511105</th>\n",
       "      <th>GSM1511106</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 44</td>\n",
       "      <td>age: 59</td>\n",
       "      <td>age: 44</td>\n",
       "      <td>age: 39</td>\n",
       "      <td>age: 64</td>\n",
       "      <td>age: 58</td>\n",
       "      <td>age: 45</td>\n",
       "      <td>age: 37</td>\n",
       "      <td>age: 40</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 58</td>\n",
       "      <td>age: 58</td>\n",
       "      <td>age: 57</td>\n",
       "      <td>age: 56</td>\n",
       "      <td>age: 54</td>\n",
       "      <td>age: 59</td>\n",
       "      <td>age: 57</td>\n",
       "      <td>age: 56</td>\n",
       "      <td>age: 56</td>\n",
       "      <td>age: 37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>...</td>\n",
       "      <td>body mass index: 39.5</td>\n",
       "      <td>body mass index: 24.9</td>\n",
       "      <td>body mass index: 29.2</td>\n",
       "      <td>body mass index: 22.2</td>\n",
       "      <td>body mass index: 26.6</td>\n",
       "      <td>body mass index: 22.1</td>\n",
       "      <td>body mass index: 28.8</td>\n",
       "      <td>body mass index: 23.9</td>\n",
       "      <td>body mass index: 27.1</td>\n",
       "      <td>body mass index: 20.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>body mass index: 22.2</td>\n",
       "      <td>body mass index: 33.1</td>\n",
       "      <td>body mass index: 22.4</td>\n",
       "      <td>body mass index: 20.6</td>\n",
       "      <td>body mass index: 27.5</td>\n",
       "      <td>body mass index: 21.9</td>\n",
       "      <td>body mass index: 26.1</td>\n",
       "      <td>body mass index: 34.8</td>\n",
       "      <td>body mass index: 20.8</td>\n",
       "      <td>...</td>\n",
       "      <td>ethnicity: AFR</td>\n",
       "      <td>ethnicity: AFR</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: AFR</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>ethnicity: AFR</td>\n",
       "      <td>ethnicity: CAU</td>\n",
       "      <td>...</td>\n",
       "      <td>gad7 score: 6</td>\n",
       "      <td>gad7 score: 1</td>\n",
       "      <td>gad7 score: 3</td>\n",
       "      <td>gad7 score: 8</td>\n",
       "      <td>gad7 score: 4</td>\n",
       "      <td>gad7 score: 7</td>\n",
       "      <td>gad7 score: 12</td>\n",
       "      <td>gad7 score: 7</td>\n",
       "      <td>gad7 score: 4</td>\n",
       "      <td>gad7 score: 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gad7 score: 2</td>\n",
       "      <td>gad7 score: 0</td>\n",
       "      <td>gad7 score: 3</td>\n",
       "      <td>gad7 score: 0</td>\n",
       "      <td>gad7 score: 7</td>\n",
       "      <td>gad7 score: 0</td>\n",
       "      <td>gad7 score: 3</td>\n",
       "      <td>gad7 score: 3</td>\n",
       "      <td>gad7 score: 4</td>\n",
       "      <td>...</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "      <td>anxiety case/control: control</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>anxiety case/control: control</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>anxiety case/control: control</td>\n",
       "      <td>anxiety case/control: case</td>\n",
       "      <td>anxiety case/control: control</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>...</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>rin: 9</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>rin: 9.6</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "      <td>rin: 9.4</td>\n",
       "      <td>hybridization batch: D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>rin: 8.1</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>rin: 7.9</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>hybridization batch: Z</td>\n",
       "      <td>rin: 6.6</td>\n",
       "      <td>rin: 7.3</td>\n",
       "      <td>rin: 6.6</td>\n",
       "      <td>...</td>\n",
       "      <td>rin: 9.4</td>\n",
       "      <td>rin: 9.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rin: 9.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rin: 9.7</td>\n",
       "      <td>rin: 9.8</td>\n",
       "      <td>rin: 9.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rin: 9.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rin: 7.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rin: 8.1</td>\n",
       "      <td>rin: 8.1</td>\n",
       "      <td>rin: 6.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã— 547 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession              GSM1510561  \\\n",
       "0  !Sample_characteristics_ch1                 age: 44   \n",
       "1  !Sample_characteristics_ch1                  Sex: F   \n",
       "2  !Sample_characteristics_ch1   body mass index: 22.2   \n",
       "3  !Sample_characteristics_ch1          ethnicity: CAU   \n",
       "4  !Sample_characteristics_ch1           gad7 score: 2   \n",
       "5  !Sample_characteristics_ch1  hybridization batch: Z   \n",
       "6  !Sample_characteristics_ch1                rin: 8.1   \n",
       "7  !Sample_characteristics_ch1                     NaN   \n",
       "\n",
       "                      GSM1510562              GSM1510563  \\\n",
       "0                        age: 59                 age: 44   \n",
       "1                         Sex: F                  Sex: F   \n",
       "2          body mass index: 33.1   body mass index: 22.4   \n",
       "3                 ethnicity: AFR          ethnicity: CAU   \n",
       "4                  gad7 score: 0           gad7 score: 3   \n",
       "5  anxiety case/control: control  hybridization batch: Z   \n",
       "6         hybridization batch: Z                rin: 7.9   \n",
       "7                       rin: 7.8                     NaN   \n",
       "\n",
       "                      GSM1510564                  GSM1510565  \\\n",
       "0                        age: 39                     age: 64   \n",
       "1                         Sex: F                      Sex: F   \n",
       "2          body mass index: 20.6       body mass index: 27.5   \n",
       "3                 ethnicity: CAU              ethnicity: CAU   \n",
       "4                  gad7 score: 0               gad7 score: 7   \n",
       "5  anxiety case/control: control  anxiety case/control: case   \n",
       "6         hybridization batch: Z      hybridization batch: Z   \n",
       "7                       rin: 8.1                    rin: 8.1   \n",
       "\n",
       "                      GSM1510566              GSM1510567  \\\n",
       "0                        age: 58                 age: 45   \n",
       "1                         Sex: M                  Sex: M   \n",
       "2          body mass index: 21.9   body mass index: 26.1   \n",
       "3                 ethnicity: CAU          ethnicity: CAU   \n",
       "4                  gad7 score: 0           gad7 score: 3   \n",
       "5  anxiety case/control: control  hybridization batch: Z   \n",
       "6         hybridization batch: Z                rin: 6.6   \n",
       "7                       rin: 6.6                     NaN   \n",
       "\n",
       "               GSM1510568              GSM1510569  ...  \\\n",
       "0                 age: 37                 age: 40  ...   \n",
       "1                  Sex: M                  Sex: M  ...   \n",
       "2   body mass index: 34.8   body mass index: 20.8  ...   \n",
       "3          ethnicity: AFR          ethnicity: CAU  ...   \n",
       "4           gad7 score: 3           gad7 score: 4  ...   \n",
       "5  hybridization batch: Z  hybridization batch: Z  ...   \n",
       "6                rin: 7.3                rin: 6.6  ...   \n",
       "7                     NaN                     NaN  ...   \n",
       "\n",
       "                   GSM1511097                     GSM1511098  \\\n",
       "0                     age: 58                        age: 58   \n",
       "1       body mass index: 39.5          body mass index: 24.9   \n",
       "2              ethnicity: AFR                 ethnicity: AFR   \n",
       "3               gad7 score: 6                  gad7 score: 1   \n",
       "4  anxiety case/control: case  anxiety case/control: control   \n",
       "5      hybridization batch: D         hybridization batch: D   \n",
       "6                    rin: 9.4                       rin: 9.7   \n",
       "7                         NaN                            NaN   \n",
       "\n",
       "               GSM1511099                  GSM1511100              GSM1511101  \\\n",
       "0                 age: 57                     age: 56                 age: 54   \n",
       "1   body mass index: 29.2       body mass index: 22.2   body mass index: 26.6   \n",
       "2          ethnicity: CAU              ethnicity: CAU          ethnicity: CAU   \n",
       "3           gad7 score: 3               gad7 score: 8           gad7 score: 4   \n",
       "4  hybridization batch: D  anxiety case/control: case  hybridization batch: D   \n",
       "5                  rin: 9      hybridization batch: D                rin: 9.6   \n",
       "6                     NaN                    rin: 9.1                     NaN   \n",
       "7                     NaN                         NaN                     NaN   \n",
       "\n",
       "                   GSM1511102                  GSM1511103  \\\n",
       "0                     age: 59                     age: 57   \n",
       "1       body mass index: 22.1       body mass index: 28.8   \n",
       "2              ethnicity: CAU              ethnicity: CAU   \n",
       "3               gad7 score: 7              gad7 score: 12   \n",
       "4  anxiety case/control: case  anxiety case/control: case   \n",
       "5      hybridization batch: D      hybridization batch: D   \n",
       "6                    rin: 9.7                    rin: 9.8   \n",
       "7                         NaN                         NaN   \n",
       "\n",
       "                   GSM1511104              GSM1511105  \\\n",
       "0                     age: 56                 age: 56   \n",
       "1       body mass index: 23.9   body mass index: 27.1   \n",
       "2              ethnicity: CAU          ethnicity: CAU   \n",
       "3               gad7 score: 7           gad7 score: 4   \n",
       "4  anxiety case/control: case  hybridization batch: D   \n",
       "5      hybridization batch: D                rin: 9.4   \n",
       "6                    rin: 9.5                     NaN   \n",
       "7                         NaN                     NaN   \n",
       "\n",
       "                   GSM1511106  \n",
       "0                     age: 37  \n",
       "1       body mass index: 20.4  \n",
       "2              ethnicity: CAU  \n",
       "3               gad7 score: 6  \n",
       "4  anxiety case/control: case  \n",
       "5      hybridization batch: D  \n",
       "6                    rin: 9.4  \n",
       "7                         NaN  \n",
       "\n",
       "[8 rows x 547 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished \n",
    "cohort = accession_num = \"GSE61672\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'gad7 score: 2', 'gad7 score: 0',\n",
       "       'gad7 score: 3', 'gad7 score: 7', 'gad7 score: 4', 'gad7 score: 9',\n",
       "       'gad7 score: 1', 'gad7 score: 10', 'gad7 score: 5',\n",
       "       'gad7 score: 17', 'gad7 score: 6', 'gad7 score: 8',\n",
       "       'gad7 score: 12', 'gad7 score: 11', 'gad7 score: 14',\n",
       "       'gad7 score: .', 'hybridization batch: Z', 'gad7 score: 18',\n",
       "       'hybridization batch: O', 'gad7 score: 13', 'gad7 score: 15',\n",
       "       'gad7 score: 20', 'gad7 score: 21', 'gad7 score: 19',\n",
       "       'anxiety case/control: case', 'anxiety case/control: control',\n",
       "       'hybridization batch: B', nan, 'hybridization batch: C',\n",
       "       'hybridization batch: D'], dtype=object)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[4]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 4\n",
    "age_row = 0\n",
    "gender_row = 1\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if (tissue_type == 'gad7 score: 2' or 'gad7 score: 3' or 'gad7 score: 4' or 'gad7 score: 1' or 'anxiety case/control: case' or 'anxiety case/control: control'):\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "    \n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1510561</th>\n",
       "      <th>GSM1510562</th>\n",
       "      <th>GSM1510563</th>\n",
       "      <th>GSM1510564</th>\n",
       "      <th>GSM1510565</th>\n",
       "      <th>GSM1510566</th>\n",
       "      <th>GSM1510567</th>\n",
       "      <th>GSM1510568</th>\n",
       "      <th>GSM1510569</th>\n",
       "      <th>GSM1510570</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1511097</th>\n",
       "      <th>GSM1511098</th>\n",
       "      <th>GSM1511099</th>\n",
       "      <th>GSM1511100</th>\n",
       "      <th>GSM1511101</th>\n",
       "      <th>GSM1511102</th>\n",
       "      <th>GSM1511103</th>\n",
       "      <th>GSM1511104</th>\n",
       "      <th>GSM1511105</th>\n",
       "      <th>GSM1511106</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>44</td>\n",
       "      <td>59</td>\n",
       "      <td>44</td>\n",
       "      <td>39</td>\n",
       "      <td>64</td>\n",
       "      <td>58</td>\n",
       "      <td>45</td>\n",
       "      <td>37</td>\n",
       "      <td>40</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "      <td>57</td>\n",
       "      <td>56</td>\n",
       "      <td>54</td>\n",
       "      <td>59</td>\n",
       "      <td>57</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 546 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM1510561  GSM1510562  GSM1510563  GSM1510564  GSM1510565  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Age                       44          59          44          39          64   \n",
       "Gender                     1           1           1           1           1   \n",
       "\n",
       "                  GSM1510566  GSM1510567  GSM1510568  GSM1510569  GSM1510570  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Age                       58          45          37          40          39   \n",
       "Gender                     0           0           0           0           1   \n",
       "\n",
       "                  ...  GSM1511097  GSM1511098  GSM1511099  GSM1511100  \\\n",
       "Anxiety disorder  ...           1           1           1           1   \n",
       "Age               ...          58          58          57          56   \n",
       "Gender            ...        None        None        None        None   \n",
       "\n",
       "                  GSM1511101  GSM1511102  GSM1511103  GSM1511104  GSM1511105  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Age                       54          59          57          56          56   \n",
       "Gender                  None        None        None        None        None   \n",
       "\n",
       "                  GSM1511106  \n",
       "Anxiety disorder           1  \n",
       "Age                       37  \n",
       "Gender                  None  \n",
       "\n",
       "[3 rows x 546 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1510561</th>\n",
       "      <th>GSM1510562</th>\n",
       "      <th>GSM1510563</th>\n",
       "      <th>GSM1510564</th>\n",
       "      <th>GSM1510565</th>\n",
       "      <th>GSM1510566</th>\n",
       "      <th>GSM1510567</th>\n",
       "      <th>GSM1510568</th>\n",
       "      <th>GSM1510569</th>\n",
       "      <th>GSM1510570</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1511097</th>\n",
       "      <th>GSM1511098</th>\n",
       "      <th>GSM1511099</th>\n",
       "      <th>GSM1511100</th>\n",
       "      <th>GSM1511101</th>\n",
       "      <th>GSM1511102</th>\n",
       "      <th>GSM1511103</th>\n",
       "      <th>GSM1511104</th>\n",
       "      <th>GSM1511105</th>\n",
       "      <th>GSM1511106</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>14.808</td>\n",
       "      <td>14.763</td>\n",
       "      <td>14.774</td>\n",
       "      <td>14.835</td>\n",
       "      <td>14.535</td>\n",
       "      <td>14.614</td>\n",
       "      <td>14.704</td>\n",
       "      <td>14.710</td>\n",
       "      <td>14.804</td>\n",
       "      <td>14.864</td>\n",
       "      <td>...</td>\n",
       "      <td>14.705</td>\n",
       "      <td>14.665</td>\n",
       "      <td>14.497</td>\n",
       "      <td>14.610</td>\n",
       "      <td>14.774</td>\n",
       "      <td>14.790</td>\n",
       "      <td>14.751</td>\n",
       "      <td>14.630</td>\n",
       "      <td>14.668</td>\n",
       "      <td>14.583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1343295</th>\n",
       "      <td>9.518</td>\n",
       "      <td>9.386</td>\n",
       "      <td>9.254</td>\n",
       "      <td>9.774</td>\n",
       "      <td>9.406</td>\n",
       "      <td>9.295</td>\n",
       "      <td>9.527</td>\n",
       "      <td>9.304</td>\n",
       "      <td>9.177</td>\n",
       "      <td>9.303</td>\n",
       "      <td>...</td>\n",
       "      <td>8.862</td>\n",
       "      <td>9.346</td>\n",
       "      <td>8.962</td>\n",
       "      <td>9.040</td>\n",
       "      <td>8.982</td>\n",
       "      <td>9.450</td>\n",
       "      <td>8.852</td>\n",
       "      <td>8.930</td>\n",
       "      <td>9.035</td>\n",
       "      <td>8.847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651228</th>\n",
       "      <td>11.092</td>\n",
       "      <td>11.183</td>\n",
       "      <td>10.725</td>\n",
       "      <td>11.114</td>\n",
       "      <td>10.845</td>\n",
       "      <td>10.658</td>\n",
       "      <td>11.262</td>\n",
       "      <td>11.182</td>\n",
       "      <td>11.011</td>\n",
       "      <td>11.466</td>\n",
       "      <td>...</td>\n",
       "      <td>12.031</td>\n",
       "      <td>11.951</td>\n",
       "      <td>11.716</td>\n",
       "      <td>12.017</td>\n",
       "      <td>12.302</td>\n",
       "      <td>11.970</td>\n",
       "      <td>11.917</td>\n",
       "      <td>12.106</td>\n",
       "      <td>12.102</td>\n",
       "      <td>12.138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651229</th>\n",
       "      <td>6.451</td>\n",
       "      <td>6.321</td>\n",
       "      <td>6.434</td>\n",
       "      <td>6.502</td>\n",
       "      <td>6.655</td>\n",
       "      <td>6.618</td>\n",
       "      <td>6.666</td>\n",
       "      <td>6.409</td>\n",
       "      <td>6.364</td>\n",
       "      <td>6.391</td>\n",
       "      <td>...</td>\n",
       "      <td>6.523</td>\n",
       "      <td>6.662</td>\n",
       "      <td>6.649</td>\n",
       "      <td>6.466</td>\n",
       "      <td>6.686</td>\n",
       "      <td>6.609</td>\n",
       "      <td>6.529</td>\n",
       "      <td>6.416</td>\n",
       "      <td>6.303</td>\n",
       "      <td>6.484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651254</th>\n",
       "      <td>8.407</td>\n",
       "      <td>8.242</td>\n",
       "      <td>8.190</td>\n",
       "      <td>8.203</td>\n",
       "      <td>8.631</td>\n",
       "      <td>8.831</td>\n",
       "      <td>8.525</td>\n",
       "      <td>8.238</td>\n",
       "      <td>8.501</td>\n",
       "      <td>8.189</td>\n",
       "      <td>...</td>\n",
       "      <td>8.633</td>\n",
       "      <td>8.625</td>\n",
       "      <td>8.288</td>\n",
       "      <td>8.253</td>\n",
       "      <td>8.224</td>\n",
       "      <td>8.571</td>\n",
       "      <td>8.228</td>\n",
       "      <td>8.230</td>\n",
       "      <td>8.217</td>\n",
       "      <td>8.247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415898</th>\n",
       "      <td>6.892</td>\n",
       "      <td>7.133</td>\n",
       "      <td>7.111</td>\n",
       "      <td>6.744</td>\n",
       "      <td>7.353</td>\n",
       "      <td>6.868</td>\n",
       "      <td>7.049</td>\n",
       "      <td>7.100</td>\n",
       "      <td>6.962</td>\n",
       "      <td>7.145</td>\n",
       "      <td>...</td>\n",
       "      <td>7.091</td>\n",
       "      <td>7.592</td>\n",
       "      <td>7.302</td>\n",
       "      <td>7.331</td>\n",
       "      <td>7.106</td>\n",
       "      <td>7.321</td>\n",
       "      <td>7.388</td>\n",
       "      <td>7.422</td>\n",
       "      <td>7.440</td>\n",
       "      <td>7.300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415911</th>\n",
       "      <td>5.529</td>\n",
       "      <td>5.597</td>\n",
       "      <td>5.637</td>\n",
       "      <td>5.653</td>\n",
       "      <td>5.714</td>\n",
       "      <td>5.578</td>\n",
       "      <td>5.613</td>\n",
       "      <td>5.850</td>\n",
       "      <td>5.849</td>\n",
       "      <td>5.779</td>\n",
       "      <td>...</td>\n",
       "      <td>5.690</td>\n",
       "      <td>5.525</td>\n",
       "      <td>5.547</td>\n",
       "      <td>5.623</td>\n",
       "      <td>5.666</td>\n",
       "      <td>5.631</td>\n",
       "      <td>5.580</td>\n",
       "      <td>5.802</td>\n",
       "      <td>5.649</td>\n",
       "      <td>5.655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415926</th>\n",
       "      <td>5.388</td>\n",
       "      <td>5.371</td>\n",
       "      <td>5.280</td>\n",
       "      <td>5.871</td>\n",
       "      <td>5.533</td>\n",
       "      <td>5.319</td>\n",
       "      <td>5.286</td>\n",
       "      <td>5.447</td>\n",
       "      <td>5.326</td>\n",
       "      <td>5.714</td>\n",
       "      <td>...</td>\n",
       "      <td>5.259</td>\n",
       "      <td>5.458</td>\n",
       "      <td>5.192</td>\n",
       "      <td>4.991</td>\n",
       "      <td>5.199</td>\n",
       "      <td>5.009</td>\n",
       "      <td>5.299</td>\n",
       "      <td>5.108</td>\n",
       "      <td>5.194</td>\n",
       "      <td>5.255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415949</th>\n",
       "      <td>6.907</td>\n",
       "      <td>6.995</td>\n",
       "      <td>6.738</td>\n",
       "      <td>7.140</td>\n",
       "      <td>6.872</td>\n",
       "      <td>6.813</td>\n",
       "      <td>7.040</td>\n",
       "      <td>6.790</td>\n",
       "      <td>6.908</td>\n",
       "      <td>6.730</td>\n",
       "      <td>...</td>\n",
       "      <td>6.857</td>\n",
       "      <td>6.740</td>\n",
       "      <td>6.866</td>\n",
       "      <td>6.781</td>\n",
       "      <td>6.892</td>\n",
       "      <td>6.756</td>\n",
       "      <td>6.846</td>\n",
       "      <td>6.924</td>\n",
       "      <td>6.849</td>\n",
       "      <td>6.757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_2415979</th>\n",
       "      <td>11.607</td>\n",
       "      <td>11.632</td>\n",
       "      <td>12.111</td>\n",
       "      <td>11.812</td>\n",
       "      <td>11.964</td>\n",
       "      <td>12.482</td>\n",
       "      <td>12.461</td>\n",
       "      <td>11.770</td>\n",
       "      <td>11.700</td>\n",
       "      <td>10.994</td>\n",
       "      <td>...</td>\n",
       "      <td>11.717</td>\n",
       "      <td>12.454</td>\n",
       "      <td>12.596</td>\n",
       "      <td>12.713</td>\n",
       "      <td>12.388</td>\n",
       "      <td>12.311</td>\n",
       "      <td>12.313</td>\n",
       "      <td>11.605</td>\n",
       "      <td>11.967</td>\n",
       "      <td>12.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12603 rows Ã— 546 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM1510561  GSM1510562  GSM1510563  GSM1510564  GSM1510565  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291      14.808      14.763      14.774      14.835      14.535   \n",
       "ILMN_1343295       9.518       9.386       9.254       9.774       9.406   \n",
       "ILMN_1651228      11.092      11.183      10.725      11.114      10.845   \n",
       "ILMN_1651229       6.451       6.321       6.434       6.502       6.655   \n",
       "ILMN_1651254       8.407       8.242       8.190       8.203       8.631   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_2415898       6.892       7.133       7.111       6.744       7.353   \n",
       "ILMN_2415911       5.529       5.597       5.637       5.653       5.714   \n",
       "ILMN_2415926       5.388       5.371       5.280       5.871       5.533   \n",
       "ILMN_2415949       6.907       6.995       6.738       7.140       6.872   \n",
       "ILMN_2415979      11.607      11.632      12.111      11.812      11.964   \n",
       "\n",
       "              GSM1510566  GSM1510567  GSM1510568  GSM1510569  GSM1510570  ...  \\\n",
       "ID                                                                        ...   \n",
       "ILMN_1343291      14.614      14.704      14.710      14.804      14.864  ...   \n",
       "ILMN_1343295       9.295       9.527       9.304       9.177       9.303  ...   \n",
       "ILMN_1651228      10.658      11.262      11.182      11.011      11.466  ...   \n",
       "ILMN_1651229       6.618       6.666       6.409       6.364       6.391  ...   \n",
       "ILMN_1651254       8.831       8.525       8.238       8.501       8.189  ...   \n",
       "...                  ...         ...         ...         ...         ...  ...   \n",
       "ILMN_2415898       6.868       7.049       7.100       6.962       7.145  ...   \n",
       "ILMN_2415911       5.578       5.613       5.850       5.849       5.779  ...   \n",
       "ILMN_2415926       5.319       5.286       5.447       5.326       5.714  ...   \n",
       "ILMN_2415949       6.813       7.040       6.790       6.908       6.730  ...   \n",
       "ILMN_2415979      12.482      12.461      11.770      11.700      10.994  ...   \n",
       "\n",
       "              GSM1511097  GSM1511098  GSM1511099  GSM1511100  GSM1511101  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291      14.705      14.665      14.497      14.610      14.774   \n",
       "ILMN_1343295       8.862       9.346       8.962       9.040       8.982   \n",
       "ILMN_1651228      12.031      11.951      11.716      12.017      12.302   \n",
       "ILMN_1651229       6.523       6.662       6.649       6.466       6.686   \n",
       "ILMN_1651254       8.633       8.625       8.288       8.253       8.224   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_2415898       7.091       7.592       7.302       7.331       7.106   \n",
       "ILMN_2415911       5.690       5.525       5.547       5.623       5.666   \n",
       "ILMN_2415926       5.259       5.458       5.192       4.991       5.199   \n",
       "ILMN_2415949       6.857       6.740       6.866       6.781       6.892   \n",
       "ILMN_2415979      11.717      12.454      12.596      12.713      12.388   \n",
       "\n",
       "              GSM1511102  GSM1511103  GSM1511104  GSM1511105  GSM1511106  \n",
       "ID                                                                        \n",
       "ILMN_1343291      14.790      14.751      14.630      14.668      14.583  \n",
       "ILMN_1343295       9.450       8.852       8.930       9.035       8.847  \n",
       "ILMN_1651228      11.970      11.917      12.106      12.102      12.138  \n",
       "ILMN_1651229       6.609       6.529       6.416       6.303       6.484  \n",
       "ILMN_1651254       8.571       8.228       8.230       8.217       8.247  \n",
       "...                  ...         ...         ...         ...         ...  \n",
       "ILMN_2415898       7.321       7.388       7.422       7.440       7.300  \n",
       "ILMN_2415911       5.631       5.580       5.802       5.649       5.655  \n",
       "ILMN_2415926       5.009       5.299       5.108       5.194       5.255  \n",
       "ILMN_2415949       6.756       6.846       6.924       6.849       6.757  \n",
       "ILMN_2415979      12.311      12.313      11.605      11.967      12.001  \n",
       "\n",
       "[12603 rows x 546 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651254', 'ILMN_1651262', 'ILMN_1651278', 'ILMN_1651296', 'ILMN_1651315', 'ILMN_1651316', 'ILMN_1651336', 'ILMN_1651346', 'ILMN_1651347', 'ILMN_1651354', 'ILMN_1651358', 'ILMN_1651364', 'ILMN_1651378', 'ILMN_1651385', 'ILMN_1651403', 'ILMN_1651405']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Species', 'Source', 'Search_Key', 'Transcript', 'ILMN_Gene',\n",
       "       'Source_Reference_ID', 'RefSeq_ID', 'Unigene_ID', 'Entrez_Gene_ID',\n",
       "       'GI', 'Accession', 'Symbol', 'Protein_Product', 'Probe_Id',\n",
       "       'Array_Address_Id', 'Probe_Type', 'Probe_Start', 'SEQUENCE',\n",
       "       'Chromosome', 'Probe_Chr_Orientation', 'Probe_Coordinates', 'Cytoband',\n",
       "       'Definition', 'Ontology_Component', 'Ontology_Process',\n",
       "       'Ontology_Function', 'Synonyms', 'Obsolete_Probe_Id', 'GB_ACC'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1510561</th>\n",
       "      <th>GSM1510562</th>\n",
       "      <th>GSM1510563</th>\n",
       "      <th>GSM1510564</th>\n",
       "      <th>GSM1510565</th>\n",
       "      <th>GSM1510566</th>\n",
       "      <th>GSM1510567</th>\n",
       "      <th>GSM1510568</th>\n",
       "      <th>GSM1510569</th>\n",
       "      <th>GSM1510570</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1511097</th>\n",
       "      <th>GSM1511098</th>\n",
       "      <th>GSM1511099</th>\n",
       "      <th>GSM1511100</th>\n",
       "      <th>GSM1511101</th>\n",
       "      <th>GSM1511102</th>\n",
       "      <th>GSM1511103</th>\n",
       "      <th>GSM1511104</th>\n",
       "      <th>GSM1511105</th>\n",
       "      <th>GSM1511106</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>6.3000</td>\n",
       "      <td>6.2350</td>\n",
       "      <td>6.4810</td>\n",
       "      <td>6.3210</td>\n",
       "      <td>6.2640</td>\n",
       "      <td>6.3570</td>\n",
       "      <td>6.2440</td>\n",
       "      <td>6.2300</td>\n",
       "      <td>6.1600</td>\n",
       "      <td>6.2880</td>\n",
       "      <td>...</td>\n",
       "      <td>6.1430</td>\n",
       "      <td>6.0480</td>\n",
       "      <td>6.0820</td>\n",
       "      <td>6.1060</td>\n",
       "      <td>6.0790</td>\n",
       "      <td>6.1280</td>\n",
       "      <td>6.1210</td>\n",
       "      <td>6.0010</td>\n",
       "      <td>6.139</td>\n",
       "      <td>6.129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>7.9120</td>\n",
       "      <td>8.0340</td>\n",
       "      <td>7.6650</td>\n",
       "      <td>7.4960</td>\n",
       "      <td>8.0450</td>\n",
       "      <td>7.6370</td>\n",
       "      <td>8.3660</td>\n",
       "      <td>7.9090</td>\n",
       "      <td>7.5420</td>\n",
       "      <td>7.5680</td>\n",
       "      <td>...</td>\n",
       "      <td>7.5470</td>\n",
       "      <td>8.0470</td>\n",
       "      <td>7.7170</td>\n",
       "      <td>7.5850</td>\n",
       "      <td>7.3850</td>\n",
       "      <td>7.8180</td>\n",
       "      <td>7.7600</td>\n",
       "      <td>7.7860</td>\n",
       "      <td>7.761</td>\n",
       "      <td>7.806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>6.7600</td>\n",
       "      <td>6.6760</td>\n",
       "      <td>6.5100</td>\n",
       "      <td>6.7020</td>\n",
       "      <td>6.7100</td>\n",
       "      <td>6.5860</td>\n",
       "      <td>6.1300</td>\n",
       "      <td>6.5600</td>\n",
       "      <td>6.7190</td>\n",
       "      <td>6.4530</td>\n",
       "      <td>...</td>\n",
       "      <td>6.6090</td>\n",
       "      <td>6.9330</td>\n",
       "      <td>6.5270</td>\n",
       "      <td>6.3300</td>\n",
       "      <td>6.4850</td>\n",
       "      <td>6.2460</td>\n",
       "      <td>6.6000</td>\n",
       "      <td>6.4350</td>\n",
       "      <td>6.349</td>\n",
       "      <td>6.570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS2</th>\n",
       "      <td>5.9320</td>\n",
       "      <td>6.1460</td>\n",
       "      <td>6.1720</td>\n",
       "      <td>6.1280</td>\n",
       "      <td>6.3120</td>\n",
       "      <td>6.0410</td>\n",
       "      <td>6.3970</td>\n",
       "      <td>6.0420</td>\n",
       "      <td>6.1950</td>\n",
       "      <td>6.0040</td>\n",
       "      <td>...</td>\n",
       "      <td>5.9640</td>\n",
       "      <td>5.7230</td>\n",
       "      <td>6.1090</td>\n",
       "      <td>5.8720</td>\n",
       "      <td>5.9690</td>\n",
       "      <td>5.8380</td>\n",
       "      <td>5.8000</td>\n",
       "      <td>6.0040</td>\n",
       "      <td>5.923</td>\n",
       "      <td>5.911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARSD1</th>\n",
       "      <td>6.5340</td>\n",
       "      <td>6.7350</td>\n",
       "      <td>6.5230</td>\n",
       "      <td>6.6050</td>\n",
       "      <td>6.9780</td>\n",
       "      <td>6.4210</td>\n",
       "      <td>6.6610</td>\n",
       "      <td>6.4630</td>\n",
       "      <td>6.1730</td>\n",
       "      <td>6.1030</td>\n",
       "      <td>...</td>\n",
       "      <td>6.1590</td>\n",
       "      <td>6.2260</td>\n",
       "      <td>6.2600</td>\n",
       "      <td>6.4460</td>\n",
       "      <td>6.2760</td>\n",
       "      <td>6.2790</td>\n",
       "      <td>6.3230</td>\n",
       "      <td>6.3910</td>\n",
       "      <td>6.152</td>\n",
       "      <td>6.246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>6.0660</td>\n",
       "      <td>5.8830</td>\n",
       "      <td>6.2500</td>\n",
       "      <td>5.7180</td>\n",
       "      <td>5.6890</td>\n",
       "      <td>6.0330</td>\n",
       "      <td>5.8980</td>\n",
       "      <td>5.9980</td>\n",
       "      <td>6.1380</td>\n",
       "      <td>5.8640</td>\n",
       "      <td>...</td>\n",
       "      <td>5.8450</td>\n",
       "      <td>5.6210</td>\n",
       "      <td>5.8320</td>\n",
       "      <td>5.8400</td>\n",
       "      <td>5.7070</td>\n",
       "      <td>5.6790</td>\n",
       "      <td>5.7950</td>\n",
       "      <td>5.8330</td>\n",
       "      <td>5.805</td>\n",
       "      <td>5.582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>7.6940</td>\n",
       "      <td>7.1850</td>\n",
       "      <td>7.5830</td>\n",
       "      <td>7.2500</td>\n",
       "      <td>7.1850</td>\n",
       "      <td>7.6110</td>\n",
       "      <td>7.9210</td>\n",
       "      <td>7.4210</td>\n",
       "      <td>7.3520</td>\n",
       "      <td>7.5220</td>\n",
       "      <td>...</td>\n",
       "      <td>7.4310</td>\n",
       "      <td>7.4150</td>\n",
       "      <td>7.4470</td>\n",
       "      <td>7.8880</td>\n",
       "      <td>7.4550</td>\n",
       "      <td>7.7730</td>\n",
       "      <td>7.7200</td>\n",
       "      <td>7.5820</td>\n",
       "      <td>7.616</td>\n",
       "      <td>7.887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>9.4940</td>\n",
       "      <td>9.1505</td>\n",
       "      <td>9.1545</td>\n",
       "      <td>9.2205</td>\n",
       "      <td>8.9365</td>\n",
       "      <td>9.1875</td>\n",
       "      <td>9.5075</td>\n",
       "      <td>9.3245</td>\n",
       "      <td>9.2635</td>\n",
       "      <td>9.1475</td>\n",
       "      <td>...</td>\n",
       "      <td>9.2445</td>\n",
       "      <td>9.3425</td>\n",
       "      <td>9.3440</td>\n",
       "      <td>9.6660</td>\n",
       "      <td>9.1590</td>\n",
       "      <td>9.3025</td>\n",
       "      <td>9.3550</td>\n",
       "      <td>9.1835</td>\n",
       "      <td>9.101</td>\n",
       "      <td>9.120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>6.5570</td>\n",
       "      <td>6.7780</td>\n",
       "      <td>6.4080</td>\n",
       "      <td>6.6600</td>\n",
       "      <td>6.4880</td>\n",
       "      <td>6.6580</td>\n",
       "      <td>6.6810</td>\n",
       "      <td>6.5490</td>\n",
       "      <td>6.6980</td>\n",
       "      <td>6.8570</td>\n",
       "      <td>...</td>\n",
       "      <td>6.6000</td>\n",
       "      <td>6.4810</td>\n",
       "      <td>6.1780</td>\n",
       "      <td>6.4850</td>\n",
       "      <td>6.5820</td>\n",
       "      <td>6.4330</td>\n",
       "      <td>6.2890</td>\n",
       "      <td>6.3560</td>\n",
       "      <td>6.402</td>\n",
       "      <td>6.717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>5.9875</td>\n",
       "      <td>6.5270</td>\n",
       "      <td>6.1310</td>\n",
       "      <td>6.2400</td>\n",
       "      <td>6.4080</td>\n",
       "      <td>6.2765</td>\n",
       "      <td>6.5245</td>\n",
       "      <td>6.2785</td>\n",
       "      <td>6.3060</td>\n",
       "      <td>6.6600</td>\n",
       "      <td>...</td>\n",
       "      <td>6.5175</td>\n",
       "      <td>6.4625</td>\n",
       "      <td>6.1285</td>\n",
       "      <td>6.2965</td>\n",
       "      <td>6.5185</td>\n",
       "      <td>6.3600</td>\n",
       "      <td>6.2085</td>\n",
       "      <td>6.4780</td>\n",
       "      <td>6.416</td>\n",
       "      <td>6.569</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7799 rows Ã— 546 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM1510561  GSM1510562  GSM1510563  GSM1510564  GSM1510565  \\\n",
       "AACS        6.3000      6.2350      6.4810      6.3210      6.2640   \n",
       "AAK1        7.9120      8.0340      7.6650      7.4960      8.0450   \n",
       "AAMP        6.7600      6.6760      6.5100      6.7020      6.7100   \n",
       "AARS2       5.9320      6.1460      6.1720      6.1280      6.3120   \n",
       "AARSD1      6.5340      6.7350      6.5230      6.6050      6.9780   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZXDC        6.0660      5.8830      6.2500      5.7180      5.6890   \n",
       "ZYG11B      7.6940      7.1850      7.5830      7.2500      7.1850   \n",
       "ZYX         9.4940      9.1505      9.1545      9.2205      8.9365   \n",
       "ZZEF1       6.5570      6.7780      6.4080      6.6600      6.4880   \n",
       "ZZZ3        5.9875      6.5270      6.1310      6.2400      6.4080   \n",
       "\n",
       "        GSM1510566  GSM1510567  GSM1510568  GSM1510569  GSM1510570  ...  \\\n",
       "AACS        6.3570      6.2440      6.2300      6.1600      6.2880  ...   \n",
       "AAK1        7.6370      8.3660      7.9090      7.5420      7.5680  ...   \n",
       "AAMP        6.5860      6.1300      6.5600      6.7190      6.4530  ...   \n",
       "AARS2       6.0410      6.3970      6.0420      6.1950      6.0040  ...   \n",
       "AARSD1      6.4210      6.6610      6.4630      6.1730      6.1030  ...   \n",
       "...            ...         ...         ...         ...         ...  ...   \n",
       "ZXDC        6.0330      5.8980      5.9980      6.1380      5.8640  ...   \n",
       "ZYG11B      7.6110      7.9210      7.4210      7.3520      7.5220  ...   \n",
       "ZYX         9.1875      9.5075      9.3245      9.2635      9.1475  ...   \n",
       "ZZEF1       6.6580      6.6810      6.5490      6.6980      6.8570  ...   \n",
       "ZZZ3        6.2765      6.5245      6.2785      6.3060      6.6600  ...   \n",
       "\n",
       "        GSM1511097  GSM1511098  GSM1511099  GSM1511100  GSM1511101  \\\n",
       "AACS        6.1430      6.0480      6.0820      6.1060      6.0790   \n",
       "AAK1        7.5470      8.0470      7.7170      7.5850      7.3850   \n",
       "AAMP        6.6090      6.9330      6.5270      6.3300      6.4850   \n",
       "AARS2       5.9640      5.7230      6.1090      5.8720      5.9690   \n",
       "AARSD1      6.1590      6.2260      6.2600      6.4460      6.2760   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "ZXDC        5.8450      5.6210      5.8320      5.8400      5.7070   \n",
       "ZYG11B      7.4310      7.4150      7.4470      7.8880      7.4550   \n",
       "ZYX         9.2445      9.3425      9.3440      9.6660      9.1590   \n",
       "ZZEF1       6.6000      6.4810      6.1780      6.4850      6.5820   \n",
       "ZZZ3        6.5175      6.4625      6.1285      6.2965      6.5185   \n",
       "\n",
       "        GSM1511102  GSM1511103  GSM1511104  GSM1511105  GSM1511106  \n",
       "AACS        6.1280      6.1210      6.0010       6.139       6.129  \n",
       "AAK1        7.8180      7.7600      7.7860       7.761       7.806  \n",
       "AAMP        6.2460      6.6000      6.4350       6.349       6.570  \n",
       "AARS2       5.8380      5.8000      6.0040       5.923       5.911  \n",
       "AARSD1      6.2790      6.3230      6.3910       6.152       6.246  \n",
       "...            ...         ...         ...         ...         ...  \n",
       "ZXDC        5.6790      5.7950      5.8330       5.805       5.582  \n",
       "ZYG11B      7.7730      7.7200      7.5820       7.616       7.887  \n",
       "ZYX         9.3025      9.3550      9.1835       9.101       9.120  \n",
       "ZZEF1       6.4330      6.2890      6.3560       6.402       6.717  \n",
       "ZZZ3        6.3600      6.2085      6.4780       6.416       6.569  \n",
       "\n",
       "[7799 rows x 546 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>AARS2</th>\n",
       "      <th>AARSD1</th>\n",
       "      <th>AASDH</th>\n",
       "      <th>AASDHPPT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM4</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1510561</th>\n",
       "      <td>1.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.3</td>\n",
       "      <td>7.912</td>\n",
       "      <td>6.76</td>\n",
       "      <td>5.932</td>\n",
       "      <td>6.534</td>\n",
       "      <td>5.799</td>\n",
       "      <td>6.122</td>\n",
       "      <td>...</td>\n",
       "      <td>6.8215</td>\n",
       "      <td>7.775</td>\n",
       "      <td>5.7</td>\n",
       "      <td>5.966</td>\n",
       "      <td>5.689</td>\n",
       "      <td>6.066</td>\n",
       "      <td>7.694</td>\n",
       "      <td>9.494</td>\n",
       "      <td>6.557</td>\n",
       "      <td>5.9875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510562</th>\n",
       "      <td>1.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.235</td>\n",
       "      <td>8.034</td>\n",
       "      <td>6.676</td>\n",
       "      <td>6.146</td>\n",
       "      <td>6.735</td>\n",
       "      <td>6.143</td>\n",
       "      <td>6.319</td>\n",
       "      <td>...</td>\n",
       "      <td>6.7145</td>\n",
       "      <td>7.276</td>\n",
       "      <td>5.773</td>\n",
       "      <td>6.297</td>\n",
       "      <td>5.765</td>\n",
       "      <td>5.883</td>\n",
       "      <td>7.185</td>\n",
       "      <td>9.1505</td>\n",
       "      <td>6.778</td>\n",
       "      <td>6.527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510563</th>\n",
       "      <td>1.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.481</td>\n",
       "      <td>7.665</td>\n",
       "      <td>6.51</td>\n",
       "      <td>6.172</td>\n",
       "      <td>6.523</td>\n",
       "      <td>5.64</td>\n",
       "      <td>6.065</td>\n",
       "      <td>...</td>\n",
       "      <td>6.656</td>\n",
       "      <td>7.648</td>\n",
       "      <td>5.527</td>\n",
       "      <td>6.334</td>\n",
       "      <td>5.725</td>\n",
       "      <td>6.25</td>\n",
       "      <td>7.583</td>\n",
       "      <td>9.1545</td>\n",
       "      <td>6.408</td>\n",
       "      <td>6.131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510564</th>\n",
       "      <td>1.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.321</td>\n",
       "      <td>7.496</td>\n",
       "      <td>6.702</td>\n",
       "      <td>6.128</td>\n",
       "      <td>6.605</td>\n",
       "      <td>5.923</td>\n",
       "      <td>6.194</td>\n",
       "      <td>...</td>\n",
       "      <td>6.641</td>\n",
       "      <td>7.205</td>\n",
       "      <td>5.814</td>\n",
       "      <td>6.709</td>\n",
       "      <td>5.774</td>\n",
       "      <td>5.718</td>\n",
       "      <td>7.25</td>\n",
       "      <td>9.2205</td>\n",
       "      <td>6.66</td>\n",
       "      <td>6.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510565</th>\n",
       "      <td>1.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.264</td>\n",
       "      <td>8.045</td>\n",
       "      <td>6.71</td>\n",
       "      <td>6.312</td>\n",
       "      <td>6.978</td>\n",
       "      <td>6.267</td>\n",
       "      <td>6.364</td>\n",
       "      <td>...</td>\n",
       "      <td>6.6235</td>\n",
       "      <td>7.285</td>\n",
       "      <td>5.833</td>\n",
       "      <td>6.437</td>\n",
       "      <td>5.678</td>\n",
       "      <td>5.689</td>\n",
       "      <td>7.185</td>\n",
       "      <td>8.9365</td>\n",
       "      <td>6.488</td>\n",
       "      <td>6.408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510892</th>\n",
       "      <td>1.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.165</td>\n",
       "      <td>7.446</td>\n",
       "      <td>6.516</td>\n",
       "      <td>6.024</td>\n",
       "      <td>6.276</td>\n",
       "      <td>5.923</td>\n",
       "      <td>6.304</td>\n",
       "      <td>...</td>\n",
       "      <td>6.7275</td>\n",
       "      <td>7.342</td>\n",
       "      <td>5.707</td>\n",
       "      <td>6.173</td>\n",
       "      <td>5.776</td>\n",
       "      <td>5.889</td>\n",
       "      <td>7.292</td>\n",
       "      <td>9.207</td>\n",
       "      <td>6.507</td>\n",
       "      <td>6.241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510893</th>\n",
       "      <td>1.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.252</td>\n",
       "      <td>7.416</td>\n",
       "      <td>6.394</td>\n",
       "      <td>5.984</td>\n",
       "      <td>6.173</td>\n",
       "      <td>5.934</td>\n",
       "      <td>6.2085</td>\n",
       "      <td>...</td>\n",
       "      <td>6.672</td>\n",
       "      <td>7.301</td>\n",
       "      <td>5.604</td>\n",
       "      <td>6.092</td>\n",
       "      <td>5.802</td>\n",
       "      <td>5.849</td>\n",
       "      <td>7.494</td>\n",
       "      <td>8.9835</td>\n",
       "      <td>6.432</td>\n",
       "      <td>6.1945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510894</th>\n",
       "      <td>1.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.227</td>\n",
       "      <td>7.42</td>\n",
       "      <td>6.451</td>\n",
       "      <td>6.083</td>\n",
       "      <td>6.311</td>\n",
       "      <td>5.871</td>\n",
       "      <td>6.27</td>\n",
       "      <td>...</td>\n",
       "      <td>6.7825</td>\n",
       "      <td>7.387</td>\n",
       "      <td>5.46</td>\n",
       "      <td>6.046</td>\n",
       "      <td>5.753</td>\n",
       "      <td>5.801</td>\n",
       "      <td>7.392</td>\n",
       "      <td>9.092</td>\n",
       "      <td>6.411</td>\n",
       "      <td>6.429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510895</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.232</td>\n",
       "      <td>7.406</td>\n",
       "      <td>6.446</td>\n",
       "      <td>5.87</td>\n",
       "      <td>6.263</td>\n",
       "      <td>5.785</td>\n",
       "      <td>6.2825</td>\n",
       "      <td>...</td>\n",
       "      <td>6.663</td>\n",
       "      <td>7.478</td>\n",
       "      <td>5.646</td>\n",
       "      <td>6.08</td>\n",
       "      <td>5.694</td>\n",
       "      <td>5.884</td>\n",
       "      <td>7.536</td>\n",
       "      <td>9.6855</td>\n",
       "      <td>6.601</td>\n",
       "      <td>6.262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1510896</th>\n",
       "      <td>1.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.152</td>\n",
       "      <td>7.529</td>\n",
       "      <td>6.512</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.24</td>\n",
       "      <td>5.884</td>\n",
       "      <td>6.1265</td>\n",
       "      <td>...</td>\n",
       "      <td>6.7125</td>\n",
       "      <td>7.46</td>\n",
       "      <td>5.51</td>\n",
       "      <td>6.067</td>\n",
       "      <td>5.711</td>\n",
       "      <td>5.937</td>\n",
       "      <td>7.486</td>\n",
       "      <td>9.5685</td>\n",
       "      <td>6.591</td>\n",
       "      <td>6.3195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>336 rows Ã— 7802 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Anxiety disorder   Age Gender   AACS   AAK1   AAMP  AARS2 AARSD1  \\\n",
       "GSM1510561              1.0  44.0    1.0    6.3  7.912   6.76  5.932  6.534   \n",
       "GSM1510562              1.0  59.0    1.0  6.235  8.034  6.676  6.146  6.735   \n",
       "GSM1510563              1.0  44.0    1.0  6.481  7.665   6.51  6.172  6.523   \n",
       "GSM1510564              1.0  39.0    1.0  6.321  7.496  6.702  6.128  6.605   \n",
       "GSM1510565              1.0  64.0    1.0  6.264  8.045   6.71  6.312  6.978   \n",
       "...                     ...   ...    ...    ...    ...    ...    ...    ...   \n",
       "GSM1510892              1.0  39.0    1.0  6.165  7.446  6.516  6.024  6.276   \n",
       "GSM1510893              1.0  60.0    1.0  6.252  7.416  6.394  5.984  6.173   \n",
       "GSM1510894              1.0  54.0    0.0  6.227   7.42  6.451  6.083  6.311   \n",
       "GSM1510895              1.0  50.0    0.0  6.232  7.406  6.446   5.87  6.263   \n",
       "GSM1510896              1.0  31.0    1.0  6.152  7.529  6.512   6.01   6.24   \n",
       "\n",
       "            AASDH AASDHPPT  ...  ZSWIM4 ZSWIM6   ZW10 ZWILCH   ZXDB   ZXDC  \\\n",
       "GSM1510561  5.799    6.122  ...  6.8215  7.775    5.7  5.966  5.689  6.066   \n",
       "GSM1510562  6.143    6.319  ...  6.7145  7.276  5.773  6.297  5.765  5.883   \n",
       "GSM1510563   5.64    6.065  ...   6.656  7.648  5.527  6.334  5.725   6.25   \n",
       "GSM1510564  5.923    6.194  ...   6.641  7.205  5.814  6.709  5.774  5.718   \n",
       "GSM1510565  6.267    6.364  ...  6.6235  7.285  5.833  6.437  5.678  5.689   \n",
       "...           ...      ...  ...     ...    ...    ...    ...    ...    ...   \n",
       "GSM1510892  5.923    6.304  ...  6.7275  7.342  5.707  6.173  5.776  5.889   \n",
       "GSM1510893  5.934   6.2085  ...   6.672  7.301  5.604  6.092  5.802  5.849   \n",
       "GSM1510894  5.871     6.27  ...  6.7825  7.387   5.46  6.046  5.753  5.801   \n",
       "GSM1510895  5.785   6.2825  ...   6.663  7.478  5.646   6.08  5.694  5.884   \n",
       "GSM1510896  5.884   6.1265  ...  6.7125   7.46   5.51  6.067  5.711  5.937   \n",
       "\n",
       "           ZYG11B     ZYX  ZZEF1    ZZZ3  \n",
       "GSM1510561  7.694   9.494  6.557  5.9875  \n",
       "GSM1510562  7.185  9.1505  6.778   6.527  \n",
       "GSM1510563  7.583  9.1545  6.408   6.131  \n",
       "GSM1510564   7.25  9.2205   6.66    6.24  \n",
       "GSM1510565  7.185  8.9365  6.488   6.408  \n",
       "...           ...     ...    ...     ...  \n",
       "GSM1510892  7.292   9.207  6.507   6.241  \n",
       "GSM1510893  7.494  8.9835  6.432  6.1945  \n",
       "GSM1510894  7.392   9.092  6.411   6.429  \n",
       "GSM1510895  7.536  9.6855  6.601   6.262  \n",
       "GSM1510896  7.486  9.5685  6.591  6.3195  \n",
       "\n",
       "[336 rows x 7802 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 336 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1.0' with 336 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 43.0\n",
      "  50% (Median): 50.0\n",
      "  75%: 56.25\n",
      "Min: 18.0\n",
      "Max: 78.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 112 occurrences. This represents 33.33% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Exposure-induced changes of plasma mRNA expression levels in patients with panic disorder\"\n",
      "!Series_summary\t\"Anxiety disorders including panic disorders with or without agoraphobia are the most prevalent mental disorders. Exposure is a core technique within the framework of cognitive behavioral therapy to treat phobia- and anxiety-related symptoms. The primary aim of this study was to trace specific anxiety-related plasma gene expression changes of subjects with PD at three time points in order to identify biomarkers for acute anxiety states. In this intervention, the patient is exposed to highly feared and mostly avoided situations.\"\n",
      "!Series_overall_design\t\"Blood samples from individuals with panic disorder (n=24) were drawn at three time points during exposure: baseline, 1 hour post-exposure and 24 hours after exposure-onset.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM3391438</th>\n",
       "      <th>GSM3391439</th>\n",
       "      <th>GSM3391440</th>\n",
       "      <th>GSM3391441</th>\n",
       "      <th>GSM3391442</th>\n",
       "      <th>GSM3391443</th>\n",
       "      <th>GSM3391444</th>\n",
       "      <th>GSM3391445</th>\n",
       "      <th>GSM3391446</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3391500</th>\n",
       "      <th>GSM3391501</th>\n",
       "      <th>GSM3391502</th>\n",
       "      <th>GSM3391503</th>\n",
       "      <th>GSM3391504</th>\n",
       "      <th>GSM3391505</th>\n",
       "      <th>GSM3391506</th>\n",
       "      <th>GSM3391507</th>\n",
       "      <th>GSM3391508</th>\n",
       "      <th>GSM3391509</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>...</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "      <td>disease: panic disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "      <td>tissue: blood plasma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: not determined</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: not determined</td>\n",
       "      <td>Sex: not determined</td>\n",
       "      <td>Sex: not determined</td>\n",
       "      <td>Sex: not determined</td>\n",
       "      <td>Sex: not determined</td>\n",
       "      <td>Sex: not determined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>medication: 1</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 0</td>\n",
       "      <td>medication: 1</td>\n",
       "      <td>medication: 1</td>\n",
       "      <td>medication: 1</td>\n",
       "      <td>medication: 1</td>\n",
       "      <td>medication: 1</td>\n",
       "      <td>medication: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>timepoint: b1</td>\n",
       "      <td>timepoint: p24_1</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: p24_1</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: b1</td>\n",
       "      <td>timepoint: b1</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: p24_1</td>\n",
       "      <td>...</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: b1</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: p24_1</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: b1</td>\n",
       "      <td>timepoint: p24_1</td>\n",
       "      <td>timepoint: pe1</td>\n",
       "      <td>timepoint: p24_1</td>\n",
       "      <td>timepoint: b1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>individual: 2</td>\n",
       "      <td>individual: 2</td>\n",
       "      <td>individual: 2</td>\n",
       "      <td>individual: 9</td>\n",
       "      <td>individual: 9</td>\n",
       "      <td>individual: 9</td>\n",
       "      <td>individual: 7</td>\n",
       "      <td>individual: 7</td>\n",
       "      <td>individual: 7</td>\n",
       "      <td>...</td>\n",
       "      <td>individual: 38</td>\n",
       "      <td>individual: 21</td>\n",
       "      <td>individual: 21</td>\n",
       "      <td>individual: 21</td>\n",
       "      <td>individual: 39</td>\n",
       "      <td>individual: 39</td>\n",
       "      <td>individual: 39</td>\n",
       "      <td>individual: 41</td>\n",
       "      <td>individual: 41</td>\n",
       "      <td>individual: 41</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession               GSM3391438  \\\n",
       "0  !Sample_characteristics_ch1  disease: panic disorder   \n",
       "1  !Sample_characteristics_ch1     tissue: blood plasma   \n",
       "2  !Sample_characteristics_ch1              Sex: female   \n",
       "3  !Sample_characteristics_ch1            medication: 0   \n",
       "4  !Sample_characteristics_ch1            timepoint: b1   \n",
       "5  !Sample_characteristics_ch1            individual: 2   \n",
       "\n",
       "                GSM3391439               GSM3391440               GSM3391441  \\\n",
       "0  disease: panic disorder  disease: panic disorder  disease: panic disorder   \n",
       "1     tissue: blood plasma     tissue: blood plasma     tissue: blood plasma   \n",
       "2              Sex: female              Sex: female              Sex: female   \n",
       "3            medication: 0            medication: 0            medication: 0   \n",
       "4         timepoint: p24_1           timepoint: pe1         timepoint: p24_1   \n",
       "5            individual: 2            individual: 2            individual: 9   \n",
       "\n",
       "                GSM3391442               GSM3391443               GSM3391444  \\\n",
       "0  disease: panic disorder  disease: panic disorder  disease: panic disorder   \n",
       "1     tissue: blood plasma     tissue: blood plasma     tissue: blood plasma   \n",
       "2              Sex: female              Sex: female                Sex: male   \n",
       "3            medication: 0            medication: 0            medication: 0   \n",
       "4           timepoint: pe1            timepoint: b1            timepoint: b1   \n",
       "5            individual: 9            individual: 9            individual: 7   \n",
       "\n",
       "                GSM3391445               GSM3391446  ...  \\\n",
       "0  disease: panic disorder  disease: panic disorder  ...   \n",
       "1     tissue: blood plasma     tissue: blood plasma  ...   \n",
       "2                Sex: male                Sex: male  ...   \n",
       "3            medication: 0            medication: 0  ...   \n",
       "4           timepoint: pe1         timepoint: p24_1  ...   \n",
       "5            individual: 7            individual: 7  ...   \n",
       "\n",
       "                GSM3391500               GSM3391501               GSM3391502  \\\n",
       "0  disease: panic disorder  disease: panic disorder  disease: panic disorder   \n",
       "1     tissue: blood plasma     tissue: blood plasma     tissue: blood plasma   \n",
       "2      Sex: not determined              Sex: female              Sex: female   \n",
       "3            medication: 1            medication: 0            medication: 0   \n",
       "4           timepoint: pe1            timepoint: b1           timepoint: pe1   \n",
       "5           individual: 38           individual: 21           individual: 21   \n",
       "\n",
       "                GSM3391503               GSM3391504               GSM3391505  \\\n",
       "0  disease: panic disorder  disease: panic disorder  disease: panic disorder   \n",
       "1     tissue: blood plasma     tissue: blood plasma     tissue: blood plasma   \n",
       "2              Sex: female      Sex: not determined      Sex: not determined   \n",
       "3            medication: 0            medication: 1            medication: 1   \n",
       "4         timepoint: p24_1           timepoint: pe1            timepoint: b1   \n",
       "5           individual: 21           individual: 39           individual: 39   \n",
       "\n",
       "                GSM3391506               GSM3391507               GSM3391508  \\\n",
       "0  disease: panic disorder  disease: panic disorder  disease: panic disorder   \n",
       "1     tissue: blood plasma     tissue: blood plasma     tissue: blood plasma   \n",
       "2      Sex: not determined      Sex: not determined      Sex: not determined   \n",
       "3            medication: 1            medication: 1            medication: 1   \n",
       "4         timepoint: p24_1           timepoint: pe1         timepoint: p24_1   \n",
       "5           individual: 39           individual: 41           individual: 41   \n",
       "\n",
       "                GSM3391509  \n",
       "0  disease: panic disorder  \n",
       "1     tissue: blood plasma  \n",
       "2      Sex: not determined  \n",
       "3            medication: 1  \n",
       "4            timepoint: b1  \n",
       "5           individual: 41  \n",
       "\n",
       "[6 rows x 73 columns]"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE119995\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease: panic disorder'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = 2\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease: panic disorder':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3391438</th>\n",
       "      <th>GSM3391439</th>\n",
       "      <th>GSM3391440</th>\n",
       "      <th>GSM3391441</th>\n",
       "      <th>GSM3391442</th>\n",
       "      <th>GSM3391443</th>\n",
       "      <th>GSM3391444</th>\n",
       "      <th>GSM3391445</th>\n",
       "      <th>GSM3391446</th>\n",
       "      <th>GSM3391447</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3391500</th>\n",
       "      <th>GSM3391501</th>\n",
       "      <th>GSM3391502</th>\n",
       "      <th>GSM3391503</th>\n",
       "      <th>GSM3391504</th>\n",
       "      <th>GSM3391505</th>\n",
       "      <th>GSM3391506</th>\n",
       "      <th>GSM3391507</th>\n",
       "      <th>GSM3391508</th>\n",
       "      <th>GSM3391509</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM3391438  GSM3391439  GSM3391440  GSM3391441  GSM3391442  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Gender                     1           1           1           1           1   \n",
       "\n",
       "                  GSM3391443  GSM3391444  GSM3391445  GSM3391446  GSM3391447  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Gender                     1           0           0           0           1   \n",
       "\n",
       "                  ...  GSM3391500  GSM3391501  GSM3391502  GSM3391503  \\\n",
       "Anxiety disorder  ...           1           1           1           1   \n",
       "Gender            ...        None           1           1           1   \n",
       "\n",
       "                  GSM3391504  GSM3391505  GSM3391506  GSM3391507  GSM3391508  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Gender                  None        None        None        None        None   \n",
       "\n",
       "                  GSM3391509  \n",
       "Anxiety disorder           1  \n",
       "Gender                  None  \n",
       "\n",
       "[2 rows x 72 columns]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3391438</th>\n",
       "      <th>GSM3391439</th>\n",
       "      <th>GSM3391440</th>\n",
       "      <th>GSM3391441</th>\n",
       "      <th>GSM3391442</th>\n",
       "      <th>GSM3391443</th>\n",
       "      <th>GSM3391444</th>\n",
       "      <th>GSM3391445</th>\n",
       "      <th>GSM3391446</th>\n",
       "      <th>GSM3391447</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3391500</th>\n",
       "      <th>GSM3391501</th>\n",
       "      <th>GSM3391502</th>\n",
       "      <th>GSM3391503</th>\n",
       "      <th>GSM3391504</th>\n",
       "      <th>GSM3391505</th>\n",
       "      <th>GSM3391506</th>\n",
       "      <th>GSM3391507</th>\n",
       "      <th>GSM3391508</th>\n",
       "      <th>GSM3391509</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>13.951749</td>\n",
       "      <td>13.632986</td>\n",
       "      <td>13.718869</td>\n",
       "      <td>14.017101</td>\n",
       "      <td>14.325731</td>\n",
       "      <td>14.328347</td>\n",
       "      <td>14.156883</td>\n",
       "      <td>13.975844</td>\n",
       "      <td>14.355057</td>\n",
       "      <td>14.263098</td>\n",
       "      <td>...</td>\n",
       "      <td>13.452625</td>\n",
       "      <td>12.726963</td>\n",
       "      <td>13.041178</td>\n",
       "      <td>13.137704</td>\n",
       "      <td>13.001847</td>\n",
       "      <td>13.462342</td>\n",
       "      <td>12.910161</td>\n",
       "      <td>13.141218</td>\n",
       "      <td>13.247583</td>\n",
       "      <td>12.954122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1343295</th>\n",
       "      <td>11.071988</td>\n",
       "      <td>10.894695</td>\n",
       "      <td>11.358074</td>\n",
       "      <td>11.219213</td>\n",
       "      <td>11.498586</td>\n",
       "      <td>11.661600</td>\n",
       "      <td>11.992226</td>\n",
       "      <td>12.115965</td>\n",
       "      <td>12.005301</td>\n",
       "      <td>11.571721</td>\n",
       "      <td>...</td>\n",
       "      <td>11.807002</td>\n",
       "      <td>11.583282</td>\n",
       "      <td>11.843324</td>\n",
       "      <td>11.818459</td>\n",
       "      <td>11.658436</td>\n",
       "      <td>11.825321</td>\n",
       "      <td>11.409416</td>\n",
       "      <td>11.742370</td>\n",
       "      <td>11.768177</td>\n",
       "      <td>11.643327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651199</th>\n",
       "      <td>5.212670</td>\n",
       "      <td>5.482078</td>\n",
       "      <td>5.357357</td>\n",
       "      <td>5.606711</td>\n",
       "      <td>4.839159</td>\n",
       "      <td>4.821573</td>\n",
       "      <td>5.241077</td>\n",
       "      <td>5.039115</td>\n",
       "      <td>4.730284</td>\n",
       "      <td>4.675609</td>\n",
       "      <td>...</td>\n",
       "      <td>4.826455</td>\n",
       "      <td>5.074913</td>\n",
       "      <td>5.010022</td>\n",
       "      <td>5.550813</td>\n",
       "      <td>4.759479</td>\n",
       "      <td>5.001379</td>\n",
       "      <td>4.736003</td>\n",
       "      <td>4.927063</td>\n",
       "      <td>4.735439</td>\n",
       "      <td>4.896824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651209</th>\n",
       "      <td>5.710210</td>\n",
       "      <td>5.611946</td>\n",
       "      <td>5.930514</td>\n",
       "      <td>5.307133</td>\n",
       "      <td>5.391922</td>\n",
       "      <td>5.643804</td>\n",
       "      <td>5.415717</td>\n",
       "      <td>5.227683</td>\n",
       "      <td>5.307049</td>\n",
       "      <td>5.712707</td>\n",
       "      <td>...</td>\n",
       "      <td>5.528267</td>\n",
       "      <td>5.145149</td>\n",
       "      <td>5.490204</td>\n",
       "      <td>4.951263</td>\n",
       "      <td>5.827258</td>\n",
       "      <td>5.639573</td>\n",
       "      <td>5.975935</td>\n",
       "      <td>5.417391</td>\n",
       "      <td>5.218732</td>\n",
       "      <td>5.548980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651210</th>\n",
       "      <td>5.385582</td>\n",
       "      <td>5.243548</td>\n",
       "      <td>5.347599</td>\n",
       "      <td>4.957354</td>\n",
       "      <td>5.386004</td>\n",
       "      <td>5.124224</td>\n",
       "      <td>4.952177</td>\n",
       "      <td>4.693552</td>\n",
       "      <td>5.285033</td>\n",
       "      <td>5.125894</td>\n",
       "      <td>...</td>\n",
       "      <td>5.045532</td>\n",
       "      <td>4.862499</td>\n",
       "      <td>5.216427</td>\n",
       "      <td>5.084498</td>\n",
       "      <td>5.025204</td>\n",
       "      <td>5.045238</td>\n",
       "      <td>5.134669</td>\n",
       "      <td>5.030738</td>\n",
       "      <td>4.727393</td>\n",
       "      <td>4.736943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311170</th>\n",
       "      <td>5.391212</td>\n",
       "      <td>5.150167</td>\n",
       "      <td>4.968791</td>\n",
       "      <td>5.215904</td>\n",
       "      <td>4.876341</td>\n",
       "      <td>4.803347</td>\n",
       "      <td>4.772063</td>\n",
       "      <td>4.738366</td>\n",
       "      <td>4.726934</td>\n",
       "      <td>5.245909</td>\n",
       "      <td>...</td>\n",
       "      <td>5.425808</td>\n",
       "      <td>4.783792</td>\n",
       "      <td>5.465189</td>\n",
       "      <td>5.279209</td>\n",
       "      <td>4.872425</td>\n",
       "      <td>4.974837</td>\n",
       "      <td>5.462659</td>\n",
       "      <td>5.198998</td>\n",
       "      <td>5.215293</td>\n",
       "      <td>5.448491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311175</th>\n",
       "      <td>5.585412</td>\n",
       "      <td>5.642667</td>\n",
       "      <td>5.141310</td>\n",
       "      <td>5.137059</td>\n",
       "      <td>5.476385</td>\n",
       "      <td>5.103688</td>\n",
       "      <td>5.073684</td>\n",
       "      <td>5.257233</td>\n",
       "      <td>5.253794</td>\n",
       "      <td>4.717288</td>\n",
       "      <td>...</td>\n",
       "      <td>5.035802</td>\n",
       "      <td>5.373508</td>\n",
       "      <td>5.464113</td>\n",
       "      <td>5.085251</td>\n",
       "      <td>5.126689</td>\n",
       "      <td>5.093751</td>\n",
       "      <td>5.102440</td>\n",
       "      <td>5.267268</td>\n",
       "      <td>4.973795</td>\n",
       "      <td>5.219592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311180</th>\n",
       "      <td>5.662226</td>\n",
       "      <td>5.262558</td>\n",
       "      <td>5.342964</td>\n",
       "      <td>5.502404</td>\n",
       "      <td>5.329657</td>\n",
       "      <td>5.426351</td>\n",
       "      <td>5.281256</td>\n",
       "      <td>5.027729</td>\n",
       "      <td>5.411350</td>\n",
       "      <td>4.545742</td>\n",
       "      <td>...</td>\n",
       "      <td>5.359871</td>\n",
       "      <td>5.789463</td>\n",
       "      <td>5.239502</td>\n",
       "      <td>4.986382</td>\n",
       "      <td>5.340640</td>\n",
       "      <td>5.046709</td>\n",
       "      <td>5.336161</td>\n",
       "      <td>5.520774</td>\n",
       "      <td>5.237579</td>\n",
       "      <td>5.110814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311185</th>\n",
       "      <td>5.220874</td>\n",
       "      <td>5.488200</td>\n",
       "      <td>5.430758</td>\n",
       "      <td>5.099603</td>\n",
       "      <td>5.371403</td>\n",
       "      <td>5.266177</td>\n",
       "      <td>4.968190</td>\n",
       "      <td>4.920104</td>\n",
       "      <td>5.291698</td>\n",
       "      <td>5.333733</td>\n",
       "      <td>...</td>\n",
       "      <td>5.440144</td>\n",
       "      <td>5.063427</td>\n",
       "      <td>4.921783</td>\n",
       "      <td>5.496954</td>\n",
       "      <td>5.005685</td>\n",
       "      <td>5.378131</td>\n",
       "      <td>5.255420</td>\n",
       "      <td>5.402068</td>\n",
       "      <td>5.737316</td>\n",
       "      <td>5.465343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_3311190</th>\n",
       "      <td>6.161327</td>\n",
       "      <td>6.045168</td>\n",
       "      <td>6.192184</td>\n",
       "      <td>6.178353</td>\n",
       "      <td>5.858747</td>\n",
       "      <td>5.991073</td>\n",
       "      <td>5.943809</td>\n",
       "      <td>5.918388</td>\n",
       "      <td>5.744141</td>\n",
       "      <td>5.624535</td>\n",
       "      <td>...</td>\n",
       "      <td>5.446380</td>\n",
       "      <td>6.029119</td>\n",
       "      <td>5.914983</td>\n",
       "      <td>6.131900</td>\n",
       "      <td>5.792538</td>\n",
       "      <td>5.883751</td>\n",
       "      <td>5.590757</td>\n",
       "      <td>5.748350</td>\n",
       "      <td>5.612821</td>\n",
       "      <td>5.811183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47290 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM3391438  GSM3391439  GSM3391440  GSM3391441  GSM3391442  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   13.951749   13.632986   13.718869   14.017101   14.325731   \n",
       "ILMN_1343295   11.071988   10.894695   11.358074   11.219213   11.498586   \n",
       "ILMN_1651199    5.212670    5.482078    5.357357    5.606711    4.839159   \n",
       "ILMN_1651209    5.710210    5.611946    5.930514    5.307133    5.391922   \n",
       "ILMN_1651210    5.385582    5.243548    5.347599    4.957354    5.386004   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_3311170    5.391212    5.150167    4.968791    5.215904    4.876341   \n",
       "ILMN_3311175    5.585412    5.642667    5.141310    5.137059    5.476385   \n",
       "ILMN_3311180    5.662226    5.262558    5.342964    5.502404    5.329657   \n",
       "ILMN_3311185    5.220874    5.488200    5.430758    5.099603    5.371403   \n",
       "ILMN_3311190    6.161327    6.045168    6.192184    6.178353    5.858747   \n",
       "\n",
       "              GSM3391443  GSM3391444  GSM3391445  GSM3391446  GSM3391447  ...  \\\n",
       "ID                                                                        ...   \n",
       "ILMN_1343291   14.328347   14.156883   13.975844   14.355057   14.263098  ...   \n",
       "ILMN_1343295   11.661600   11.992226   12.115965   12.005301   11.571721  ...   \n",
       "ILMN_1651199    4.821573    5.241077    5.039115    4.730284    4.675609  ...   \n",
       "ILMN_1651209    5.643804    5.415717    5.227683    5.307049    5.712707  ...   \n",
       "ILMN_1651210    5.124224    4.952177    4.693552    5.285033    5.125894  ...   \n",
       "...                  ...         ...         ...         ...         ...  ...   \n",
       "ILMN_3311170    4.803347    4.772063    4.738366    4.726934    5.245909  ...   \n",
       "ILMN_3311175    5.103688    5.073684    5.257233    5.253794    4.717288  ...   \n",
       "ILMN_3311180    5.426351    5.281256    5.027729    5.411350    4.545742  ...   \n",
       "ILMN_3311185    5.266177    4.968190    4.920104    5.291698    5.333733  ...   \n",
       "ILMN_3311190    5.991073    5.943809    5.918388    5.744141    5.624535  ...   \n",
       "\n",
       "              GSM3391500  GSM3391501  GSM3391502  GSM3391503  GSM3391504  \\\n",
       "ID                                                                         \n",
       "ILMN_1343291   13.452625   12.726963   13.041178   13.137704   13.001847   \n",
       "ILMN_1343295   11.807002   11.583282   11.843324   11.818459   11.658436   \n",
       "ILMN_1651199    4.826455    5.074913    5.010022    5.550813    4.759479   \n",
       "ILMN_1651209    5.528267    5.145149    5.490204    4.951263    5.827258   \n",
       "ILMN_1651210    5.045532    4.862499    5.216427    5.084498    5.025204   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "ILMN_3311170    5.425808    4.783792    5.465189    5.279209    4.872425   \n",
       "ILMN_3311175    5.035802    5.373508    5.464113    5.085251    5.126689   \n",
       "ILMN_3311180    5.359871    5.789463    5.239502    4.986382    5.340640   \n",
       "ILMN_3311185    5.440144    5.063427    4.921783    5.496954    5.005685   \n",
       "ILMN_3311190    5.446380    6.029119    5.914983    6.131900    5.792538   \n",
       "\n",
       "              GSM3391505  GSM3391506  GSM3391507  GSM3391508  GSM3391509  \n",
       "ID                                                                        \n",
       "ILMN_1343291   13.462342   12.910161   13.141218   13.247583   12.954122  \n",
       "ILMN_1343295   11.825321   11.409416   11.742370   11.768177   11.643327  \n",
       "ILMN_1651199    5.001379    4.736003    4.927063    4.735439    4.896824  \n",
       "ILMN_1651209    5.639573    5.975935    5.417391    5.218732    5.548980  \n",
       "ILMN_1651210    5.045238    5.134669    5.030738    4.727393    4.736943  \n",
       "...                  ...         ...         ...         ...         ...  \n",
       "ILMN_3311170    4.974837    5.462659    5.198998    5.215293    5.448491  \n",
       "ILMN_3311175    5.093751    5.102440    5.267268    4.973795    5.219592  \n",
       "ILMN_3311180    5.046709    5.336161    5.520774    5.237579    5.110814  \n",
       "ILMN_3311185    5.378131    5.255420    5.402068    5.737316    5.465343  \n",
       "ILMN_3311190    5.883751    5.590757    5.748350    5.612821    5.811183  \n",
       "\n",
       "[47290 rows x 72 columns]"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651199', 'ILMN_1651209', 'ILMN_1651210', 'ILMN_1651221', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651230', 'ILMN_1651232', 'ILMN_1651235', 'ILMN_1651236', 'ILMN_1651237', 'ILMN_1651238', 'ILMN_1651249', 'ILMN_1651253', 'ILMN_1651254', 'ILMN_1651259', 'ILMN_1651260', 'ILMN_1651262']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Species', 'Source', 'Search_Key', 'Transcript', 'ILMN_Gene',\n",
       "       'Source_Reference_ID', 'RefSeq_ID', 'Unigene_ID', 'Entrez_Gene_ID',\n",
       "       'GI', 'Accession', 'Symbol', 'Protein_Product', 'Probe_Id',\n",
       "       'Array_Address_Id', 'Probe_Type', 'Probe_Start', 'SEQUENCE',\n",
       "       'Chromosome', 'Probe_Chr_Orientation', 'Probe_Coordinates', 'Cytoband',\n",
       "       'Definition', 'Ontology_Component', 'Ontology_Process',\n",
       "       'Ontology_Function', 'Synonyms', 'Obsolete_Probe_Id', 'GB_ACC'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3391438</th>\n",
       "      <th>GSM3391439</th>\n",
       "      <th>GSM3391440</th>\n",
       "      <th>GSM3391441</th>\n",
       "      <th>GSM3391442</th>\n",
       "      <th>GSM3391443</th>\n",
       "      <th>GSM3391444</th>\n",
       "      <th>GSM3391445</th>\n",
       "      <th>GSM3391446</th>\n",
       "      <th>GSM3391447</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3391500</th>\n",
       "      <th>GSM3391501</th>\n",
       "      <th>GSM3391502</th>\n",
       "      <th>GSM3391503</th>\n",
       "      <th>GSM3391504</th>\n",
       "      <th>GSM3391505</th>\n",
       "      <th>GSM3391506</th>\n",
       "      <th>GSM3391507</th>\n",
       "      <th>GSM3391508</th>\n",
       "      <th>GSM3391509</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.549873</td>\n",
       "      <td>5.156616</td>\n",
       "      <td>5.547066</td>\n",
       "      <td>5.302168</td>\n",
       "      <td>5.185350</td>\n",
       "      <td>5.186735</td>\n",
       "      <td>5.319098</td>\n",
       "      <td>5.102512</td>\n",
       "      <td>5.078470</td>\n",
       "      <td>5.539373</td>\n",
       "      <td>...</td>\n",
       "      <td>5.382586</td>\n",
       "      <td>5.576582</td>\n",
       "      <td>5.413309</td>\n",
       "      <td>5.522419</td>\n",
       "      <td>5.247449</td>\n",
       "      <td>5.161875</td>\n",
       "      <td>5.481058</td>\n",
       "      <td>5.447319</td>\n",
       "      <td>5.551500</td>\n",
       "      <td>5.466158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.236366</td>\n",
       "      <td>4.993963</td>\n",
       "      <td>5.103459</td>\n",
       "      <td>5.089547</td>\n",
       "      <td>4.955472</td>\n",
       "      <td>5.042418</td>\n",
       "      <td>4.984974</td>\n",
       "      <td>5.009426</td>\n",
       "      <td>5.085790</td>\n",
       "      <td>5.284906</td>\n",
       "      <td>...</td>\n",
       "      <td>5.368677</td>\n",
       "      <td>5.368314</td>\n",
       "      <td>5.098376</td>\n",
       "      <td>5.195656</td>\n",
       "      <td>5.073764</td>\n",
       "      <td>5.052990</td>\n",
       "      <td>4.698542</td>\n",
       "      <td>5.235947</td>\n",
       "      <td>5.367221</td>\n",
       "      <td>5.286537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>4.676289</td>\n",
       "      <td>4.170137</td>\n",
       "      <td>3.772850</td>\n",
       "      <td>3.713753</td>\n",
       "      <td>4.116266</td>\n",
       "      <td>3.496562</td>\n",
       "      <td>4.170447</td>\n",
       "      <td>4.677483</td>\n",
       "      <td>4.418204</td>\n",
       "      <td>3.900678</td>\n",
       "      <td>...</td>\n",
       "      <td>3.740830</td>\n",
       "      <td>4.377500</td>\n",
       "      <td>4.039470</td>\n",
       "      <td>3.831097</td>\n",
       "      <td>4.090791</td>\n",
       "      <td>3.959632</td>\n",
       "      <td>4.693917</td>\n",
       "      <td>4.111218</td>\n",
       "      <td>4.111810</td>\n",
       "      <td>4.263751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>4.692780</td>\n",
       "      <td>4.125878</td>\n",
       "      <td>4.425559</td>\n",
       "      <td>3.463952</td>\n",
       "      <td>4.574096</td>\n",
       "      <td>4.156001</td>\n",
       "      <td>4.954480</td>\n",
       "      <td>4.528565</td>\n",
       "      <td>4.246560</td>\n",
       "      <td>4.486061</td>\n",
       "      <td>...</td>\n",
       "      <td>4.582576</td>\n",
       "      <td>3.784288</td>\n",
       "      <td>4.655316</td>\n",
       "      <td>3.924939</td>\n",
       "      <td>4.819901</td>\n",
       "      <td>3.122133</td>\n",
       "      <td>4.649517</td>\n",
       "      <td>4.163421</td>\n",
       "      <td>4.967549</td>\n",
       "      <td>4.784429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>5.148601</td>\n",
       "      <td>5.177554</td>\n",
       "      <td>5.003242</td>\n",
       "      <td>4.878174</td>\n",
       "      <td>4.777945</td>\n",
       "      <td>5.341176</td>\n",
       "      <td>4.879487</td>\n",
       "      <td>4.935164</td>\n",
       "      <td>5.069078</td>\n",
       "      <td>5.282842</td>\n",
       "      <td>...</td>\n",
       "      <td>4.682381</td>\n",
       "      <td>5.265869</td>\n",
       "      <td>5.196862</td>\n",
       "      <td>4.947796</td>\n",
       "      <td>5.090187</td>\n",
       "      <td>5.133800</td>\n",
       "      <td>4.693469</td>\n",
       "      <td>4.802838</td>\n",
       "      <td>5.274135</td>\n",
       "      <td>5.047789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>6.771469</td>\n",
       "      <td>6.336628</td>\n",
       "      <td>6.555160</td>\n",
       "      <td>6.129595</td>\n",
       "      <td>5.900300</td>\n",
       "      <td>6.191537</td>\n",
       "      <td>5.782648</td>\n",
       "      <td>6.413790</td>\n",
       "      <td>6.128815</td>\n",
       "      <td>6.406779</td>\n",
       "      <td>...</td>\n",
       "      <td>4.913304</td>\n",
       "      <td>5.037243</td>\n",
       "      <td>5.635179</td>\n",
       "      <td>5.186833</td>\n",
       "      <td>5.426234</td>\n",
       "      <td>4.902439</td>\n",
       "      <td>5.288266</td>\n",
       "      <td>4.998581</td>\n",
       "      <td>4.986692</td>\n",
       "      <td>5.337996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>9.569284</td>\n",
       "      <td>9.326883</td>\n",
       "      <td>9.911704</td>\n",
       "      <td>9.542863</td>\n",
       "      <td>9.504159</td>\n",
       "      <td>9.546907</td>\n",
       "      <td>9.848659</td>\n",
       "      <td>10.509720</td>\n",
       "      <td>9.726760</td>\n",
       "      <td>9.400222</td>\n",
       "      <td>...</td>\n",
       "      <td>9.122883</td>\n",
       "      <td>9.403109</td>\n",
       "      <td>9.491428</td>\n",
       "      <td>9.310131</td>\n",
       "      <td>9.807796</td>\n",
       "      <td>9.505578</td>\n",
       "      <td>9.945647</td>\n",
       "      <td>9.577237</td>\n",
       "      <td>9.028118</td>\n",
       "      <td>9.320740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>6.099535</td>\n",
       "      <td>5.957916</td>\n",
       "      <td>6.237110</td>\n",
       "      <td>6.430535</td>\n",
       "      <td>6.247545</td>\n",
       "      <td>6.389494</td>\n",
       "      <td>6.482721</td>\n",
       "      <td>6.755369</td>\n",
       "      <td>6.471711</td>\n",
       "      <td>6.180198</td>\n",
       "      <td>...</td>\n",
       "      <td>5.476549</td>\n",
       "      <td>4.864323</td>\n",
       "      <td>5.268544</td>\n",
       "      <td>5.469607</td>\n",
       "      <td>6.201858</td>\n",
       "      <td>5.980847</td>\n",
       "      <td>5.804774</td>\n",
       "      <td>5.236338</td>\n",
       "      <td>5.575930</td>\n",
       "      <td>5.052478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>5.464955</td>\n",
       "      <td>5.349479</td>\n",
       "      <td>5.307907</td>\n",
       "      <td>5.278758</td>\n",
       "      <td>5.418133</td>\n",
       "      <td>5.036321</td>\n",
       "      <td>5.517764</td>\n",
       "      <td>5.098876</td>\n",
       "      <td>5.583733</td>\n",
       "      <td>5.428541</td>\n",
       "      <td>...</td>\n",
       "      <td>5.559138</td>\n",
       "      <td>5.364978</td>\n",
       "      <td>5.011174</td>\n",
       "      <td>4.990195</td>\n",
       "      <td>5.108184</td>\n",
       "      <td>5.317062</td>\n",
       "      <td>5.322555</td>\n",
       "      <td>5.345468</td>\n",
       "      <td>5.269554</td>\n",
       "      <td>4.930351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB1C</th>\n",
       "      <td>4.960658</td>\n",
       "      <td>5.100309</td>\n",
       "      <td>4.314269</td>\n",
       "      <td>4.822136</td>\n",
       "      <td>5.013245</td>\n",
       "      <td>5.487166</td>\n",
       "      <td>4.830636</td>\n",
       "      <td>4.987027</td>\n",
       "      <td>5.000327</td>\n",
       "      <td>4.818966</td>\n",
       "      <td>...</td>\n",
       "      <td>5.386724</td>\n",
       "      <td>5.240203</td>\n",
       "      <td>4.784970</td>\n",
       "      <td>4.865959</td>\n",
       "      <td>4.977681</td>\n",
       "      <td>5.033901</td>\n",
       "      <td>4.884116</td>\n",
       "      <td>4.971943</td>\n",
       "      <td>3.781470</td>\n",
       "      <td>5.006437</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17886 rows Ã— 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM3391438  GSM3391439  GSM3391440  GSM3391441  GSM3391442  \\\n",
       "A1BG       5.549873    5.156616    5.547066    5.302168    5.185350   \n",
       "A1CF       5.236366    4.993963    5.103459    5.089547    4.955472   \n",
       "A2M        4.676289    4.170137    3.772850    3.713753    4.116266   \n",
       "A2ML1      4.692780    4.125878    4.425559    3.463952    4.574096   \n",
       "A3GALT2    5.148601    5.177554    5.003242    4.878174    4.777945   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B     6.771469    6.336628    6.555160    6.129595    5.900300   \n",
       "ZYX        9.569284    9.326883    9.911704    9.542863    9.504159   \n",
       "ZZEF1      6.099535    5.957916    6.237110    6.430535    6.247545   \n",
       "ZZZ3       5.464955    5.349479    5.307907    5.278758    5.418133   \n",
       "RAB1C      4.960658    5.100309    4.314269    4.822136    5.013245   \n",
       "\n",
       "         GSM3391443  GSM3391444  GSM3391445  GSM3391446  GSM3391447  ...  \\\n",
       "A1BG       5.186735    5.319098    5.102512    5.078470    5.539373  ...   \n",
       "A1CF       5.042418    4.984974    5.009426    5.085790    5.284906  ...   \n",
       "A2M        3.496562    4.170447    4.677483    4.418204    3.900678  ...   \n",
       "A2ML1      4.156001    4.954480    4.528565    4.246560    4.486061  ...   \n",
       "A3GALT2    5.341176    4.879487    4.935164    5.069078    5.282842  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZYG11B     6.191537    5.782648    6.413790    6.128815    6.406779  ...   \n",
       "ZYX        9.546907    9.848659   10.509720    9.726760    9.400222  ...   \n",
       "ZZEF1      6.389494    6.482721    6.755369    6.471711    6.180198  ...   \n",
       "ZZZ3       5.036321    5.517764    5.098876    5.583733    5.428541  ...   \n",
       "RAB1C      5.487166    4.830636    4.987027    5.000327    4.818966  ...   \n",
       "\n",
       "         GSM3391500  GSM3391501  GSM3391502  GSM3391503  GSM3391504  \\\n",
       "A1BG       5.382586    5.576582    5.413309    5.522419    5.247449   \n",
       "A1CF       5.368677    5.368314    5.098376    5.195656    5.073764   \n",
       "A2M        3.740830    4.377500    4.039470    3.831097    4.090791   \n",
       "A2ML1      4.582576    3.784288    4.655316    3.924939    4.819901   \n",
       "A3GALT2    4.682381    5.265869    5.196862    4.947796    5.090187   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYG11B     4.913304    5.037243    5.635179    5.186833    5.426234   \n",
       "ZYX        9.122883    9.403109    9.491428    9.310131    9.807796   \n",
       "ZZEF1      5.476549    4.864323    5.268544    5.469607    6.201858   \n",
       "ZZZ3       5.559138    5.364978    5.011174    4.990195    5.108184   \n",
       "RAB1C      5.386724    5.240203    4.784970    4.865959    4.977681   \n",
       "\n",
       "         GSM3391505  GSM3391506  GSM3391507  GSM3391508  GSM3391509  \n",
       "A1BG       5.161875    5.481058    5.447319    5.551500    5.466158  \n",
       "A1CF       5.052990    4.698542    5.235947    5.367221    5.286537  \n",
       "A2M        3.959632    4.693917    4.111218    4.111810    4.263751  \n",
       "A2ML1      3.122133    4.649517    4.163421    4.967549    4.784429  \n",
       "A3GALT2    5.133800    4.693469    4.802838    5.274135    5.047789  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZYG11B     4.902439    5.288266    4.998581    4.986692    5.337996  \n",
       "ZYX        9.505578    9.945647    9.577237    9.028118    9.320740  \n",
       "ZZEF1      5.980847    5.804774    5.236338    5.575930    5.052478  \n",
       "ZZZ3       5.317062    5.322555    5.345468    5.269554    4.930351  \n",
       "RAB1C      5.033901    4.884116    4.971943    3.781470    5.006437  \n",
       "\n",
       "[17886 rows x 72 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>RAB1C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3391438</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.549873</td>\n",
       "      <td>5.236366</td>\n",
       "      <td>4.676289</td>\n",
       "      <td>4.69278</td>\n",
       "      <td>5.148601</td>\n",
       "      <td>4.803079</td>\n",
       "      <td>5.458767</td>\n",
       "      <td>4.923727</td>\n",
       "      <td>...</td>\n",
       "      <td>4.964826</td>\n",
       "      <td>4.940303</td>\n",
       "      <td>5.073779</td>\n",
       "      <td>5.884059</td>\n",
       "      <td>4.705465</td>\n",
       "      <td>6.771469</td>\n",
       "      <td>9.569284</td>\n",
       "      <td>6.099535</td>\n",
       "      <td>5.464955</td>\n",
       "      <td>4.960658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391439</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.156616</td>\n",
       "      <td>4.993963</td>\n",
       "      <td>4.170137</td>\n",
       "      <td>4.125878</td>\n",
       "      <td>5.177554</td>\n",
       "      <td>5.119226</td>\n",
       "      <td>5.515966</td>\n",
       "      <td>4.804336</td>\n",
       "      <td>...</td>\n",
       "      <td>4.753434</td>\n",
       "      <td>4.591868</td>\n",
       "      <td>5.164072</td>\n",
       "      <td>6.149183</td>\n",
       "      <td>4.846971</td>\n",
       "      <td>6.336628</td>\n",
       "      <td>9.326883</td>\n",
       "      <td>5.957916</td>\n",
       "      <td>5.349479</td>\n",
       "      <td>5.100309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391440</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.547066</td>\n",
       "      <td>5.103459</td>\n",
       "      <td>3.77285</td>\n",
       "      <td>4.425559</td>\n",
       "      <td>5.003242</td>\n",
       "      <td>4.216696</td>\n",
       "      <td>5.356336</td>\n",
       "      <td>4.647138</td>\n",
       "      <td>...</td>\n",
       "      <td>4.705461</td>\n",
       "      <td>4.46119</td>\n",
       "      <td>4.788716</td>\n",
       "      <td>6.228165</td>\n",
       "      <td>4.89554</td>\n",
       "      <td>6.55516</td>\n",
       "      <td>9.911704</td>\n",
       "      <td>6.23711</td>\n",
       "      <td>5.307907</td>\n",
       "      <td>4.314269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391441</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.302168</td>\n",
       "      <td>5.089547</td>\n",
       "      <td>3.713753</td>\n",
       "      <td>3.463952</td>\n",
       "      <td>4.878174</td>\n",
       "      <td>4.084012</td>\n",
       "      <td>5.809323</td>\n",
       "      <td>5.224995</td>\n",
       "      <td>...</td>\n",
       "      <td>4.709073</td>\n",
       "      <td>4.607347</td>\n",
       "      <td>4.687015</td>\n",
       "      <td>6.370589</td>\n",
       "      <td>5.048973</td>\n",
       "      <td>6.129595</td>\n",
       "      <td>9.542863</td>\n",
       "      <td>6.430535</td>\n",
       "      <td>5.278758</td>\n",
       "      <td>4.822136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391442</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.18535</td>\n",
       "      <td>4.955472</td>\n",
       "      <td>4.116266</td>\n",
       "      <td>4.574096</td>\n",
       "      <td>4.777945</td>\n",
       "      <td>4.557257</td>\n",
       "      <td>5.611132</td>\n",
       "      <td>4.936271</td>\n",
       "      <td>...</td>\n",
       "      <td>4.883024</td>\n",
       "      <td>4.638102</td>\n",
       "      <td>5.072173</td>\n",
       "      <td>6.126019</td>\n",
       "      <td>4.675962</td>\n",
       "      <td>5.9003</td>\n",
       "      <td>9.504159</td>\n",
       "      <td>6.247545</td>\n",
       "      <td>5.418133</td>\n",
       "      <td>5.013245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391496</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.506376</td>\n",
       "      <td>5.145666</td>\n",
       "      <td>3.867223</td>\n",
       "      <td>4.486624</td>\n",
       "      <td>5.333725</td>\n",
       "      <td>4.400585</td>\n",
       "      <td>5.74246</td>\n",
       "      <td>4.953283</td>\n",
       "      <td>...</td>\n",
       "      <td>4.669475</td>\n",
       "      <td>4.705037</td>\n",
       "      <td>5.28401</td>\n",
       "      <td>6.567619</td>\n",
       "      <td>4.826816</td>\n",
       "      <td>5.64686</td>\n",
       "      <td>10.161538</td>\n",
       "      <td>6.28309</td>\n",
       "      <td>5.190047</td>\n",
       "      <td>4.530331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391497</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.173176</td>\n",
       "      <td>5.040722</td>\n",
       "      <td>4.074068</td>\n",
       "      <td>4.549112</td>\n",
       "      <td>5.013721</td>\n",
       "      <td>4.583834</td>\n",
       "      <td>5.466677</td>\n",
       "      <td>4.70262</td>\n",
       "      <td>...</td>\n",
       "      <td>4.72745</td>\n",
       "      <td>4.840712</td>\n",
       "      <td>4.972748</td>\n",
       "      <td>6.626783</td>\n",
       "      <td>4.806269</td>\n",
       "      <td>5.501809</td>\n",
       "      <td>10.196891</td>\n",
       "      <td>6.33844</td>\n",
       "      <td>5.275097</td>\n",
       "      <td>4.90162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391501</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.576582</td>\n",
       "      <td>5.368314</td>\n",
       "      <td>4.3775</td>\n",
       "      <td>3.784288</td>\n",
       "      <td>5.265869</td>\n",
       "      <td>4.882237</td>\n",
       "      <td>6.312621</td>\n",
       "      <td>4.832429</td>\n",
       "      <td>...</td>\n",
       "      <td>4.479698</td>\n",
       "      <td>4.54106</td>\n",
       "      <td>5.122266</td>\n",
       "      <td>6.304932</td>\n",
       "      <td>4.705346</td>\n",
       "      <td>5.037243</td>\n",
       "      <td>9.403109</td>\n",
       "      <td>4.864323</td>\n",
       "      <td>5.364978</td>\n",
       "      <td>5.240203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391502</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.413309</td>\n",
       "      <td>5.098376</td>\n",
       "      <td>4.03947</td>\n",
       "      <td>4.655316</td>\n",
       "      <td>5.196862</td>\n",
       "      <td>3.692935</td>\n",
       "      <td>5.458813</td>\n",
       "      <td>4.765107</td>\n",
       "      <td>...</td>\n",
       "      <td>4.344076</td>\n",
       "      <td>4.844574</td>\n",
       "      <td>4.838627</td>\n",
       "      <td>6.203816</td>\n",
       "      <td>4.915858</td>\n",
       "      <td>5.635179</td>\n",
       "      <td>9.491428</td>\n",
       "      <td>5.268544</td>\n",
       "      <td>5.011174</td>\n",
       "      <td>4.78497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3391503</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.522419</td>\n",
       "      <td>5.195656</td>\n",
       "      <td>3.831097</td>\n",
       "      <td>3.924939</td>\n",
       "      <td>4.947796</td>\n",
       "      <td>3.902718</td>\n",
       "      <td>5.594524</td>\n",
       "      <td>4.967341</td>\n",
       "      <td>...</td>\n",
       "      <td>4.654772</td>\n",
       "      <td>4.880905</td>\n",
       "      <td>5.102385</td>\n",
       "      <td>6.322168</td>\n",
       "      <td>4.95105</td>\n",
       "      <td>5.186833</td>\n",
       "      <td>9.310131</td>\n",
       "      <td>5.469607</td>\n",
       "      <td>4.990195</td>\n",
       "      <td>4.865959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 17888 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Anxiety disorder Gender      A1BG      A1CF       A2M     A2ML1  \\\n",
       "GSM3391438              1.0    1.0  5.549873  5.236366  4.676289   4.69278   \n",
       "GSM3391439              1.0    1.0  5.156616  4.993963  4.170137  4.125878   \n",
       "GSM3391440              1.0    1.0  5.547066  5.103459   3.77285  4.425559   \n",
       "GSM3391441              1.0    1.0  5.302168  5.089547  3.713753  3.463952   \n",
       "GSM3391442              1.0    1.0   5.18535  4.955472  4.116266  4.574096   \n",
       "...                     ...    ...       ...       ...       ...       ...   \n",
       "GSM3391496              1.0    0.0  5.506376  5.145666  3.867223  4.486624   \n",
       "GSM3391497              1.0    0.0  5.173176  5.040722  4.074068  4.549112   \n",
       "GSM3391501              1.0    1.0  5.576582  5.368314    4.3775  3.784288   \n",
       "GSM3391502              1.0    1.0  5.413309  5.098376   4.03947  4.655316   \n",
       "GSM3391503              1.0    1.0  5.522419  5.195656  3.831097  3.924939   \n",
       "\n",
       "             A3GALT2    A4GALT     A4GNT      AAA1  ...     ZWINT      ZXDA  \\\n",
       "GSM3391438  5.148601  4.803079  5.458767  4.923727  ...  4.964826  4.940303   \n",
       "GSM3391439  5.177554  5.119226  5.515966  4.804336  ...  4.753434  4.591868   \n",
       "GSM3391440  5.003242  4.216696  5.356336  4.647138  ...  4.705461   4.46119   \n",
       "GSM3391441  4.878174  4.084012  5.809323  5.224995  ...  4.709073  4.607347   \n",
       "GSM3391442  4.777945  4.557257  5.611132  4.936271  ...  4.883024  4.638102   \n",
       "...              ...       ...       ...       ...  ...       ...       ...   \n",
       "GSM3391496  5.333725  4.400585   5.74246  4.953283  ...  4.669475  4.705037   \n",
       "GSM3391497  5.013721  4.583834  5.466677   4.70262  ...   4.72745  4.840712   \n",
       "GSM3391501  5.265869  4.882237  6.312621  4.832429  ...  4.479698   4.54106   \n",
       "GSM3391502  5.196862  3.692935  5.458813  4.765107  ...  4.344076  4.844574   \n",
       "GSM3391503  4.947796  3.902718  5.594524  4.967341  ...  4.654772  4.880905   \n",
       "\n",
       "                ZXDB      ZXDC    ZYG11A    ZYG11B        ZYX     ZZEF1  \\\n",
       "GSM3391438  5.073779  5.884059  4.705465  6.771469   9.569284  6.099535   \n",
       "GSM3391439  5.164072  6.149183  4.846971  6.336628   9.326883  5.957916   \n",
       "GSM3391440  4.788716  6.228165   4.89554   6.55516   9.911704   6.23711   \n",
       "GSM3391441  4.687015  6.370589  5.048973  6.129595   9.542863  6.430535   \n",
       "GSM3391442  5.072173  6.126019  4.675962    5.9003   9.504159  6.247545   \n",
       "...              ...       ...       ...       ...        ...       ...   \n",
       "GSM3391496   5.28401  6.567619  4.826816   5.64686  10.161538   6.28309   \n",
       "GSM3391497  4.972748  6.626783  4.806269  5.501809  10.196891   6.33844   \n",
       "GSM3391501  5.122266  6.304932  4.705346  5.037243   9.403109  4.864323   \n",
       "GSM3391502  4.838627  6.203816  4.915858  5.635179   9.491428  5.268544   \n",
       "GSM3391503  5.102385  6.322168   4.95105  5.186833   9.310131  5.469607   \n",
       "\n",
       "                ZZZ3     RAB1C  \n",
       "GSM3391438  5.464955  4.960658  \n",
       "GSM3391439  5.349479  5.100309  \n",
       "GSM3391440  5.307907  4.314269  \n",
       "GSM3391441  5.278758  4.822136  \n",
       "GSM3391442  5.418133  5.013245  \n",
       "...              ...       ...  \n",
       "GSM3391496  5.190047  4.530331  \n",
       "GSM3391497  5.275097   4.90162  \n",
       "GSM3391501  5.364978  5.240203  \n",
       "GSM3391502  5.011174   4.78497  \n",
       "GSM3391503  4.990195  4.865959  \n",
       "\n",
       "[63 rows x 17888 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 63 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1.0' with 63 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is severely biased.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 18 occurrences. This represents 28.57% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Blood gene expression signatures distinguish autism spectrum disorders from controls\"\n",
      "!Series_summary\t\"Autism Spectrum Disorder (ASD) is a common pediatric cognitive disorder with high heritability. Yet no single genetic variant has accounted for more than a small fraction of cases. We sought to determine whether we could classify patients as having ASD vs. controls solely based on a multi-gene expression profiling of their peripheral blood cells.\"\n",
      "!Series_overall_design\t\"To test whether peripheral blood gene expression profiles could be used as a molecular diagnostic tool for distinguishing ASD from controls, we performed peripheral blood gene expression profiling on 170 patients with ASD and 115 controls collected from Boston area hospitals.  We developed a 55-gene prediction model with a sample cohort of 66 male patients with ASD and 33 age-matched male controls using cross-validation strategy.  Subsequently, 104 ASD and 82 controls were recruited and used as first and second validation sets, respectively.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM650510</th>\n",
       "      <th>GSM650512</th>\n",
       "      <th>GSM650513</th>\n",
       "      <th>GSM650514</th>\n",
       "      <th>GSM650515</th>\n",
       "      <th>GSM650516</th>\n",
       "      <th>GSM650517</th>\n",
       "      <th>GSM650518</th>\n",
       "      <th>GSM650519</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM650644</th>\n",
       "      <th>GSM650645</th>\n",
       "      <th>GSM650646</th>\n",
       "      <th>GSM650647</th>\n",
       "      <th>GSM650648</th>\n",
       "      <th>GSM650649</th>\n",
       "      <th>GSM650651</th>\n",
       "      <th>GSM650652</th>\n",
       "      <th>GSM650653</th>\n",
       "      <th>GSM650654</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>diagnosis: PDD-NOS</td>\n",
       "      <td>diagnosis: PDD-NOS</td>\n",
       "      <td>diagnosis: AUTISM</td>\n",
       "      <td>diagnosis: AUTISM</td>\n",
       "      <td>diagnosis: AUTISM</td>\n",
       "      <td>diagnosis: PDD-NOS</td>\n",
       "      <td>diagnosis: PDD-NOS</td>\n",
       "      <td>diagnosis: AUTISM</td>\n",
       "      <td>diagnosis: ASPERGER'S DISORDER</td>\n",
       "      <td>...</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "      <td>diagnosis: CONTROL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age at blood drawing (months): 118</td>\n",
       "      <td>age at blood drawing (months): 79</td>\n",
       "      <td>age at blood drawing (months): 169</td>\n",
       "      <td>age at blood drawing (months): 92</td>\n",
       "      <td>age at blood drawing (months): 78</td>\n",
       "      <td>age at blood drawing (months): 124</td>\n",
       "      <td>age at blood drawing (months): 81</td>\n",
       "      <td>age at blood drawing (months): 82</td>\n",
       "      <td>age at blood drawing (months): 96</td>\n",
       "      <td>...</td>\n",
       "      <td>age at blood drawing (months): 157</td>\n",
       "      <td>age at blood drawing (months): 160</td>\n",
       "      <td>age at blood drawing (months): 163</td>\n",
       "      <td>age at blood drawing (months): 156</td>\n",
       "      <td>age at blood drawing (months): 167</td>\n",
       "      <td>age at blood drawing (months): 168</td>\n",
       "      <td>age at blood drawing (months): 183</td>\n",
       "      <td>age at blood drawing (months): 185</td>\n",
       "      <td>age at blood drawing (months): 191</td>\n",
       "      <td>age at blood drawing (months): 192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Mixed (Caucasian, Asian)</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>...</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Black</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: Unknown</td>\n",
       "      <td>race: Caucasian</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>...</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "      <td>ethnicity: Non-Hispanic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>hours (minutes since last caloric intake): : 0:30</td>\n",
       "      <td>hours (minutes since last caloric intake): : 0:10</td>\n",
       "      <td>hours (minutes since last caloric intake): : 1:00</td>\n",
       "      <td>hours (minutes since last caloric intake): : 1:00</td>\n",
       "      <td>hours (minutes since last caloric intake): : 0:05</td>\n",
       "      <td>hours (minutes since last caloric intake): : 2:30</td>\n",
       "      <td>hours (minutes since last caloric intake): : 2:30</td>\n",
       "      <td>hours (minutes since last caloric intake): : 8:00</td>\n",
       "      <td>hours (minutes since last caloric intake): : 1:30</td>\n",
       "      <td>...</td>\n",
       "      <td>hours (minutes since last caloric intake): : 2:30</td>\n",
       "      <td>hours (minutes since last caloric intake): : 5:10</td>\n",
       "      <td>hours (minutes since last caloric intake): : 3:00</td>\n",
       "      <td>hours (minutes since last caloric intake): : u...</td>\n",
       "      <td>hours (minutes since last caloric intake): : u...</td>\n",
       "      <td>hours (minutes since last caloric intake): : u...</td>\n",
       "      <td>hours (minutes since last caloric intake): : u...</td>\n",
       "      <td>hours (minutes since last caloric intake): : u...</td>\n",
       "      <td>hours (minutes since last caloric intake): : u...</td>\n",
       "      <td>hours (minutes since last caloric intake): : 4:10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>...</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "      <td>chronic diseases: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>...</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "      <td>allergies: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: yes (mild menta...</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: yes (mild menta...</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: yes (mild menta...</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>...</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "      <td>developmental/speech disorder: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: yes (Siezure/ convulsio...</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>...</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "      <td>neurological disorder: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>psychiatric disorder: yes (anxiety)</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: yes (add)</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: yes (adhd/add)</td>\n",
       "      <td>...</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: yes (ADHD)</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: yes (ADHD)</td>\n",
       "      <td>psychiatric disorder: -</td>\n",
       "      <td>psychiatric disorder: yes (ADD)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gastrointestinal disorder: yes (reflux)</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: yes (Ulcerative col...</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>...</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "      <td>gastrointestinal disorder: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>...</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: yes (asthma)</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "      <td>autoimmune disorder: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: 7q21 amplification</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>...</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "      <td>genetic testing: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>...</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "      <td>birth defects: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>...</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "      <td>ct: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>...</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "      <td>mri: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: Straterra</td>\n",
       "      <td>medications/vitamin names: Miralax (glycolax)</td>\n",
       "      <td>medications/vitamin names: Colazol, Prevacid</td>\n",
       "      <td>medications/vitamin names: Concerta, Straterra</td>\n",
       "      <td>...</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: Adderall</td>\n",
       "      <td>medications/vitamin names: albuterol, flovent</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: -</td>\n",
       "      <td>medications/vitamin names: concerta, Amitrypilene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>...</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "      <td>cerebral palsy: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: yes</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>...</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "      <td>seizure: -</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>...</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "      <td>other diseases: -</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows Ã— 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          !Sample_geo_accession  \\\n",
       "0   !Sample_characteristics_ch1   \n",
       "1   !Sample_characteristics_ch1   \n",
       "2   !Sample_characteristics_ch1   \n",
       "3   !Sample_characteristics_ch1   \n",
       "4   !Sample_characteristics_ch1   \n",
       "5   !Sample_characteristics_ch1   \n",
       "6   !Sample_characteristics_ch1   \n",
       "7   !Sample_characteristics_ch1   \n",
       "8   !Sample_characteristics_ch1   \n",
       "9   !Sample_characteristics_ch1   \n",
       "10  !Sample_characteristics_ch1   \n",
       "11  !Sample_characteristics_ch1   \n",
       "12  !Sample_characteristics_ch1   \n",
       "13  !Sample_characteristics_ch1   \n",
       "14  !Sample_characteristics_ch1   \n",
       "15  !Sample_characteristics_ch1   \n",
       "16  !Sample_characteristics_ch1   \n",
       "17  !Sample_characteristics_ch1   \n",
       "18  !Sample_characteristics_ch1   \n",
       "19  !Sample_characteristics_ch1   \n",
       "20  !Sample_characteristics_ch1   \n",
       "\n",
       "                                            GSM650510  \\\n",
       "0                                  diagnosis: PDD-NOS   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 118   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 0:30   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                psychiatric disorder: yes (anxiety)   \n",
       "11            gastrointestinal disorder: yes (reflux)   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650512  \\\n",
       "0                                  diagnosis: PDD-NOS   \n",
       "1                                        gender: male   \n",
       "2                   age at blood drawing (months): 79   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 0:10   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650513  \\\n",
       "0                                   diagnosis: AUTISM   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 169   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 1:00   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8   developmental/speech disorder: yes (mild menta...   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650514  \\\n",
       "0                                   diagnosis: AUTISM   \n",
       "1                                        gender: male   \n",
       "2                   age at blood drawing (months): 92   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 1:00   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650515  \\\n",
       "0                                   diagnosis: AUTISM   \n",
       "1                                        gender: male   \n",
       "2                   age at blood drawing (months): 78   \n",
       "3                      race: Mixed (Caucasian, Asian)   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 0:05   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8   developmental/speech disorder: yes (mild menta...   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650516  \\\n",
       "0                                  diagnosis: PDD-NOS   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 124   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 2:30   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                    psychiatric disorder: yes (add)   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                genetic testing: 7q21 amplification   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17               medications/vitamin names: Straterra   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650517  \\\n",
       "0                                  diagnosis: PDD-NOS   \n",
       "1                                        gender: male   \n",
       "2                   age at blood drawing (months): 81   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 2:30   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17      medications/vitamin names: Miralax (glycolax)   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650518  \\\n",
       "0                                   diagnosis: AUTISM   \n",
       "1                                        gender: male   \n",
       "2                   age at blood drawing (months): 82   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 8:00   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8   developmental/speech disorder: yes (mild menta...   \n",
       "9   neurological disorder: yes (Siezure/ convulsio...   \n",
       "10                            psychiatric disorder: -   \n",
       "11  gastrointestinal disorder: yes (Ulcerative col...   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17       medications/vitamin names: Colazol, Prevacid   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                       seizure: yes   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650519  ...  \\\n",
       "0                      diagnosis: ASPERGER'S DISORDER  ...   \n",
       "1                                        gender: male  ...   \n",
       "2                   age at blood drawing (months): 96  ...   \n",
       "3                                     race: Caucasian  ...   \n",
       "4                             ethnicity: Non-Hispanic  ...   \n",
       "5   hours (minutes since last caloric intake): : 1:30  ...   \n",
       "6                                 chronic diseases: -  ...   \n",
       "7                                        allergies: -  ...   \n",
       "8                    developmental/speech disorder: -  ...   \n",
       "9                            neurological disorder: -  ...   \n",
       "10               psychiatric disorder: yes (adhd/add)  ...   \n",
       "11                       gastrointestinal disorder: -  ...   \n",
       "12                             autoimmune disorder: -  ...   \n",
       "13                                 genetic testing: -  ...   \n",
       "14                                   birth defects: -  ...   \n",
       "15                                              ct: -  ...   \n",
       "16                                             mri: -  ...   \n",
       "17     medications/vitamin names: Concerta, Straterra  ...   \n",
       "18                                  cerebral palsy: -  ...   \n",
       "19                                         seizure: -  ...   \n",
       "20                                  other diseases: -  ...   \n",
       "\n",
       "                                            GSM650644  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 157   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 2:30   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650645  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 160   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 5:10   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                   psychiatric disorder: yes (ADHD)   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                medications/vitamin names: Adderall   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650646  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 163   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : 3:00   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                  autoimmune disorder: yes (asthma)   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17      medications/vitamin names: albuterol, flovent   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650647  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 156   \n",
       "3                                     race: Caucasian   \n",
       "4                                 ethnicity: Hispanic   \n",
       "5   hours (minutes since last caloric intake): : u...   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650648  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 167   \n",
       "3                                         race: Black   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : u...   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650649  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 168   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : u...   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650651  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 183   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : u...   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650652  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 185   \n",
       "3                                     race: Caucasian   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : u...   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                   psychiatric disorder: yes (ADHD)   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650653  \\\n",
       "0                                  diagnosis: CONTROL   \n",
       "1                                        gender: male   \n",
       "2                  age at blood drawing (months): 191   \n",
       "3                                       race: Unknown   \n",
       "4                             ethnicity: Non-Hispanic   \n",
       "5   hours (minutes since last caloric intake): : u...   \n",
       "6                                 chronic diseases: -   \n",
       "7                                        allergies: -   \n",
       "8                    developmental/speech disorder: -   \n",
       "9                            neurological disorder: -   \n",
       "10                            psychiatric disorder: -   \n",
       "11                       gastrointestinal disorder: -   \n",
       "12                             autoimmune disorder: -   \n",
       "13                                 genetic testing: -   \n",
       "14                                   birth defects: -   \n",
       "15                                              ct: -   \n",
       "16                                             mri: -   \n",
       "17                       medications/vitamin names: -   \n",
       "18                                  cerebral palsy: -   \n",
       "19                                         seizure: -   \n",
       "20                                  other diseases: -   \n",
       "\n",
       "                                            GSM650654  \n",
       "0                                  diagnosis: CONTROL  \n",
       "1                                        gender: male  \n",
       "2                  age at blood drawing (months): 192  \n",
       "3                                     race: Caucasian  \n",
       "4                             ethnicity: Non-Hispanic  \n",
       "5   hours (minutes since last caloric intake): : 4:10  \n",
       "6                                 chronic diseases: -  \n",
       "7                                        allergies: -  \n",
       "8                    developmental/speech disorder: -  \n",
       "9                            neurological disorder: -  \n",
       "10                    psychiatric disorder: yes (ADD)  \n",
       "11                       gastrointestinal disorder: -  \n",
       "12                             autoimmune disorder: -  \n",
       "13                                 genetic testing: -  \n",
       "14                                   birth defects: -  \n",
       "15                                              ct: -  \n",
       "16                                             mri: -  \n",
       "17  medications/vitamin names: concerta, Amitrypilene  \n",
       "18                                  cerebral palsy: -  \n",
       "19                                         seizure: -  \n",
       "20                                  other diseases: -  \n",
       "\n",
       "[21 rows x 100 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished \n",
    "cohort = accession_num = \"GSE18123\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'psychiatric disorder: yes (anxiety)', 'psychiatric disorder: -',\n",
       "       'psychiatric disorder: yes (add)',\n",
       "       'psychiatric disorder: yes (adhd/add)',\n",
       "       'psychiatric disorder: yes (adhd)',\n",
       "       'psychiatric disorder: yes (pica)', 'gastrointestinal disorder: -',\n",
       "       'psychiatric disorder: unknown',\n",
       "       'psychiatric disorder: yes (ADHD)',\n",
       "       'psychiatric disorder: yes (behaviorial issues)',\n",
       "       'psychiatric disorder: yes (ADD)'], dtype=object)"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[10]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 10\n",
    "age_row = 2\n",
    "gender_row = 1\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'psychiatric disorder: yes (anxiety)':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = round(int(age_string.split(': ')[1])/12, 2)\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "    \n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM650510</th>\n",
       "      <th>GSM650512</th>\n",
       "      <th>GSM650513</th>\n",
       "      <th>GSM650514</th>\n",
       "      <th>GSM650515</th>\n",
       "      <th>GSM650516</th>\n",
       "      <th>GSM650517</th>\n",
       "      <th>GSM650518</th>\n",
       "      <th>GSM650519</th>\n",
       "      <th>GSM650520</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM650644</th>\n",
       "      <th>GSM650645</th>\n",
       "      <th>GSM650646</th>\n",
       "      <th>GSM650647</th>\n",
       "      <th>GSM650648</th>\n",
       "      <th>GSM650649</th>\n",
       "      <th>GSM650651</th>\n",
       "      <th>GSM650652</th>\n",
       "      <th>GSM650653</th>\n",
       "      <th>GSM650654</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>1.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>9.83</td>\n",
       "      <td>6.58</td>\n",
       "      <td>14.08</td>\n",
       "      <td>7.67</td>\n",
       "      <td>6.5</td>\n",
       "      <td>10.33</td>\n",
       "      <td>6.75</td>\n",
       "      <td>6.83</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.75</td>\n",
       "      <td>...</td>\n",
       "      <td>13.08</td>\n",
       "      <td>13.33</td>\n",
       "      <td>13.58</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.92</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15.25</td>\n",
       "      <td>15.42</td>\n",
       "      <td>15.92</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM650510  GSM650512  GSM650513  GSM650514  GSM650515  \\\n",
       "Anxiety disorder       1.00       0.00       0.00       0.00        0.0   \n",
       "Age                    9.83       6.58      14.08       7.67        6.5   \n",
       "Gender                 0.00       0.00       0.00       0.00        0.0   \n",
       "\n",
       "                  GSM650516  GSM650517  GSM650518  GSM650519  GSM650520  ...  \\\n",
       "Anxiety disorder       0.00       0.00       0.00        0.0       0.00  ...   \n",
       "Age                   10.33       6.75       6.83        8.0       9.75  ...   \n",
       "Gender                 0.00       0.00       0.00        0.0       0.00  ...   \n",
       "\n",
       "                  GSM650644  GSM650645  GSM650646  GSM650647  GSM650648  \\\n",
       "Anxiety disorder       0.00       0.00       0.00        0.0       0.00   \n",
       "Age                   13.08      13.33      13.58       13.0      13.92   \n",
       "Gender                 0.00       0.00       0.00        0.0       0.00   \n",
       "\n",
       "                  GSM650649  GSM650651  GSM650652  GSM650653  GSM650654  \n",
       "Anxiety disorder        0.0       0.00       0.00       0.00        0.0  \n",
       "Age                    14.0      15.25      15.42      15.92       16.0  \n",
       "Gender                  0.0       0.00       0.00       0.00        0.0  \n",
       "\n",
       "[3 rows x 99 columns]"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "\n",
    "selected_clinical_data.loc['Anxiety disorder'] = selected_clinical_data.loc['Anxiety disorder'].astype(int)\n",
    "selected_clinical_data.loc['Gender'] = selected_clinical_data.loc['Gender'].astype(int)\n",
    "\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM650510</th>\n",
       "      <th>GSM650512</th>\n",
       "      <th>GSM650513</th>\n",
       "      <th>GSM650514</th>\n",
       "      <th>GSM650515</th>\n",
       "      <th>GSM650516</th>\n",
       "      <th>GSM650517</th>\n",
       "      <th>GSM650518</th>\n",
       "      <th>GSM650519</th>\n",
       "      <th>GSM650520</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM650644</th>\n",
       "      <th>GSM650645</th>\n",
       "      <th>GSM650646</th>\n",
       "      <th>GSM650647</th>\n",
       "      <th>GSM650648</th>\n",
       "      <th>GSM650649</th>\n",
       "      <th>GSM650651</th>\n",
       "      <th>GSM650652</th>\n",
       "      <th>GSM650653</th>\n",
       "      <th>GSM650654</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>26.65640</td>\n",
       "      <td>107.61809</td>\n",
       "      <td>48.97026</td>\n",
       "      <td>72.65312</td>\n",
       "      <td>80.83264</td>\n",
       "      <td>40.13933</td>\n",
       "      <td>47.43693</td>\n",
       "      <td>57.90003</td>\n",
       "      <td>65.03053</td>\n",
       "      <td>43.14078</td>\n",
       "      <td>...</td>\n",
       "      <td>81.02742</td>\n",
       "      <td>67.71790</td>\n",
       "      <td>49.50204</td>\n",
       "      <td>64.66049</td>\n",
       "      <td>110.58686</td>\n",
       "      <td>18.62053</td>\n",
       "      <td>66.15737</td>\n",
       "      <td>76.31541</td>\n",
       "      <td>59.93721</td>\n",
       "      <td>45.99937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>39.00726</td>\n",
       "      <td>135.73965</td>\n",
       "      <td>51.37555</td>\n",
       "      <td>102.65829</td>\n",
       "      <td>78.93955</td>\n",
       "      <td>81.75657</td>\n",
       "      <td>79.74300</td>\n",
       "      <td>74.16394</td>\n",
       "      <td>131.21986</td>\n",
       "      <td>62.78481</td>\n",
       "      <td>...</td>\n",
       "      <td>157.17724</td>\n",
       "      <td>86.83151</td>\n",
       "      <td>106.72321</td>\n",
       "      <td>37.18311</td>\n",
       "      <td>45.04578</td>\n",
       "      <td>77.12382</td>\n",
       "      <td>92.94504</td>\n",
       "      <td>137.79575</td>\n",
       "      <td>26.70354</td>\n",
       "      <td>101.07505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>296.01724</td>\n",
       "      <td>459.14382</td>\n",
       "      <td>378.86384</td>\n",
       "      <td>618.79730</td>\n",
       "      <td>266.87741</td>\n",
       "      <td>320.33815</td>\n",
       "      <td>685.72845</td>\n",
       "      <td>285.89160</td>\n",
       "      <td>996.76063</td>\n",
       "      <td>231.43584</td>\n",
       "      <td>...</td>\n",
       "      <td>913.09583</td>\n",
       "      <td>285.02298</td>\n",
       "      <td>539.54601</td>\n",
       "      <td>75.35680</td>\n",
       "      <td>350.91103</td>\n",
       "      <td>488.36033</td>\n",
       "      <td>252.50309</td>\n",
       "      <td>329.28063</td>\n",
       "      <td>145.30082</td>\n",
       "      <td>276.80480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>209.03575</td>\n",
       "      <td>232.08352</td>\n",
       "      <td>190.15020</td>\n",
       "      <td>185.90196</td>\n",
       "      <td>216.47953</td>\n",
       "      <td>172.25500</td>\n",
       "      <td>188.12218</td>\n",
       "      <td>155.99192</td>\n",
       "      <td>172.30327</td>\n",
       "      <td>141.44389</td>\n",
       "      <td>...</td>\n",
       "      <td>212.27610</td>\n",
       "      <td>309.88049</td>\n",
       "      <td>206.78336</td>\n",
       "      <td>281.03506</td>\n",
       "      <td>189.41136</td>\n",
       "      <td>294.80112</td>\n",
       "      <td>274.37934</td>\n",
       "      <td>232.94968</td>\n",
       "      <td>204.35111</td>\n",
       "      <td>265.61533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>63.82730</td>\n",
       "      <td>50.20000</td>\n",
       "      <td>60.23913</td>\n",
       "      <td>22.03061</td>\n",
       "      <td>41.29200</td>\n",
       "      <td>74.49414</td>\n",
       "      <td>16.84841</td>\n",
       "      <td>48.88556</td>\n",
       "      <td>19.70762</td>\n",
       "      <td>17.45545</td>\n",
       "      <td>...</td>\n",
       "      <td>42.52692</td>\n",
       "      <td>43.31678</td>\n",
       "      <td>21.37225</td>\n",
       "      <td>61.94335</td>\n",
       "      <td>37.83045</td>\n",
       "      <td>19.08806</td>\n",
       "      <td>43.59119</td>\n",
       "      <td>49.46663</td>\n",
       "      <td>36.18051</td>\n",
       "      <td>16.70463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91703_at</th>\n",
       "      <td>28.58766</td>\n",
       "      <td>79.13262</td>\n",
       "      <td>22.96121</td>\n",
       "      <td>56.42256</td>\n",
       "      <td>45.11952</td>\n",
       "      <td>34.81763</td>\n",
       "      <td>57.04181</td>\n",
       "      <td>56.23074</td>\n",
       "      <td>105.49385</td>\n",
       "      <td>41.24448</td>\n",
       "      <td>...</td>\n",
       "      <td>69.51566</td>\n",
       "      <td>37.86502</td>\n",
       "      <td>64.40379</td>\n",
       "      <td>28.92133</td>\n",
       "      <td>74.31037</td>\n",
       "      <td>35.14159</td>\n",
       "      <td>43.61235</td>\n",
       "      <td>59.13127</td>\n",
       "      <td>40.32274</td>\n",
       "      <td>38.93599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91816_f_at</th>\n",
       "      <td>0.10766</td>\n",
       "      <td>15.94046</td>\n",
       "      <td>6.61843</td>\n",
       "      <td>11.72339</td>\n",
       "      <td>7.64792</td>\n",
       "      <td>9.16214</td>\n",
       "      <td>4.53755</td>\n",
       "      <td>14.51436</td>\n",
       "      <td>20.55896</td>\n",
       "      <td>12.60243</td>\n",
       "      <td>...</td>\n",
       "      <td>13.79920</td>\n",
       "      <td>10.53937</td>\n",
       "      <td>11.12295</td>\n",
       "      <td>11.67345</td>\n",
       "      <td>10.92976</td>\n",
       "      <td>5.50120</td>\n",
       "      <td>6.32524</td>\n",
       "      <td>17.96985</td>\n",
       "      <td>4.65278</td>\n",
       "      <td>8.73903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91826_at</th>\n",
       "      <td>72.29617</td>\n",
       "      <td>42.95838</td>\n",
       "      <td>71.19351</td>\n",
       "      <td>43.94883</td>\n",
       "      <td>47.81855</td>\n",
       "      <td>54.23739</td>\n",
       "      <td>51.09696</td>\n",
       "      <td>53.49288</td>\n",
       "      <td>29.99378</td>\n",
       "      <td>77.21878</td>\n",
       "      <td>...</td>\n",
       "      <td>28.16819</td>\n",
       "      <td>66.09037</td>\n",
       "      <td>47.22587</td>\n",
       "      <td>58.17597</td>\n",
       "      <td>35.79578</td>\n",
       "      <td>61.40504</td>\n",
       "      <td>52.68096</td>\n",
       "      <td>47.94715</td>\n",
       "      <td>93.90870</td>\n",
       "      <td>45.76559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91920_at</th>\n",
       "      <td>80.84484</td>\n",
       "      <td>76.71756</td>\n",
       "      <td>125.63001</td>\n",
       "      <td>72.40296</td>\n",
       "      <td>105.65854</td>\n",
       "      <td>96.91254</td>\n",
       "      <td>80.77544</td>\n",
       "      <td>66.64779</td>\n",
       "      <td>65.19188</td>\n",
       "      <td>67.64769</td>\n",
       "      <td>...</td>\n",
       "      <td>71.52543</td>\n",
       "      <td>71.21265</td>\n",
       "      <td>74.19817</td>\n",
       "      <td>122.47701</td>\n",
       "      <td>77.78997</td>\n",
       "      <td>114.24505</td>\n",
       "      <td>122.05100</td>\n",
       "      <td>77.77357</td>\n",
       "      <td>117.51711</td>\n",
       "      <td>121.57002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91952_at</th>\n",
       "      <td>31.52882</td>\n",
       "      <td>60.39404</td>\n",
       "      <td>57.26548</td>\n",
       "      <td>74.95020</td>\n",
       "      <td>72.19887</td>\n",
       "      <td>64.50435</td>\n",
       "      <td>63.80760</td>\n",
       "      <td>76.21304</td>\n",
       "      <td>91.48289</td>\n",
       "      <td>61.29613</td>\n",
       "      <td>...</td>\n",
       "      <td>70.73734</td>\n",
       "      <td>54.79309</td>\n",
       "      <td>77.71017</td>\n",
       "      <td>68.49546</td>\n",
       "      <td>162.56649</td>\n",
       "      <td>68.31865</td>\n",
       "      <td>66.56453</td>\n",
       "      <td>85.00159</td>\n",
       "      <td>64.24315</td>\n",
       "      <td>62.66024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54613 rows Ã— 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM650510  GSM650512  GSM650513  GSM650514  GSM650515  GSM650516  \\\n",
       "ID                                                                             \n",
       "1007_s_at    26.65640  107.61809   48.97026   72.65312   80.83264   40.13933   \n",
       "1053_at      39.00726  135.73965   51.37555  102.65829   78.93955   81.75657   \n",
       "117_at      296.01724  459.14382  378.86384  618.79730  266.87741  320.33815   \n",
       "121_at      209.03575  232.08352  190.15020  185.90196  216.47953  172.25500   \n",
       "1255_g_at    63.82730   50.20000   60.23913   22.03061   41.29200   74.49414   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "91703_at     28.58766   79.13262   22.96121   56.42256   45.11952   34.81763   \n",
       "91816_f_at    0.10766   15.94046    6.61843   11.72339    7.64792    9.16214   \n",
       "91826_at     72.29617   42.95838   71.19351   43.94883   47.81855   54.23739   \n",
       "91920_at     80.84484   76.71756  125.63001   72.40296  105.65854   96.91254   \n",
       "91952_at     31.52882   60.39404   57.26548   74.95020   72.19887   64.50435   \n",
       "\n",
       "            GSM650517  GSM650518  GSM650519  GSM650520  ...  GSM650644  \\\n",
       "ID                                                      ...              \n",
       "1007_s_at    47.43693   57.90003   65.03053   43.14078  ...   81.02742   \n",
       "1053_at      79.74300   74.16394  131.21986   62.78481  ...  157.17724   \n",
       "117_at      685.72845  285.89160  996.76063  231.43584  ...  913.09583   \n",
       "121_at      188.12218  155.99192  172.30327  141.44389  ...  212.27610   \n",
       "1255_g_at    16.84841   48.88556   19.70762   17.45545  ...   42.52692   \n",
       "...               ...        ...        ...        ...  ...        ...   \n",
       "91703_at     57.04181   56.23074  105.49385   41.24448  ...   69.51566   \n",
       "91816_f_at    4.53755   14.51436   20.55896   12.60243  ...   13.79920   \n",
       "91826_at     51.09696   53.49288   29.99378   77.21878  ...   28.16819   \n",
       "91920_at     80.77544   66.64779   65.19188   67.64769  ...   71.52543   \n",
       "91952_at     63.80760   76.21304   91.48289   61.29613  ...   70.73734   \n",
       "\n",
       "            GSM650645  GSM650646  GSM650647  GSM650648  GSM650649  GSM650651  \\\n",
       "ID                                                                             \n",
       "1007_s_at    67.71790   49.50204   64.66049  110.58686   18.62053   66.15737   \n",
       "1053_at      86.83151  106.72321   37.18311   45.04578   77.12382   92.94504   \n",
       "117_at      285.02298  539.54601   75.35680  350.91103  488.36033  252.50309   \n",
       "121_at      309.88049  206.78336  281.03506  189.41136  294.80112  274.37934   \n",
       "1255_g_at    43.31678   21.37225   61.94335   37.83045   19.08806   43.59119   \n",
       "...               ...        ...        ...        ...        ...        ...   \n",
       "91703_at     37.86502   64.40379   28.92133   74.31037   35.14159   43.61235   \n",
       "91816_f_at   10.53937   11.12295   11.67345   10.92976    5.50120    6.32524   \n",
       "91826_at     66.09037   47.22587   58.17597   35.79578   61.40504   52.68096   \n",
       "91920_at     71.21265   74.19817  122.47701   77.78997  114.24505  122.05100   \n",
       "91952_at     54.79309   77.71017   68.49546  162.56649   68.31865   66.56453   \n",
       "\n",
       "            GSM650652  GSM650653  GSM650654  \n",
       "ID                                           \n",
       "1007_s_at    76.31541   59.93721   45.99937  \n",
       "1053_at     137.79575   26.70354  101.07505  \n",
       "117_at      329.28063  145.30082  276.80480  \n",
       "121_at      232.94968  204.35111  265.61533  \n",
       "1255_g_at    49.46663   36.18051   16.70463  \n",
       "...               ...        ...        ...  \n",
       "91703_at     59.13127   40.32274   38.93599  \n",
       "91816_f_at   17.96985    4.65278    8.73903  \n",
       "91826_at     47.94715   93.90870   45.76559  \n",
       "91920_at     77.77357  117.51711  121.57002  \n",
       "91952_at     85.00159   64.24315   62.66024  \n",
       "\n",
       "[54613 rows x 99 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM650510</th>\n",
       "      <th>GSM650512</th>\n",
       "      <th>GSM650513</th>\n",
       "      <th>GSM650514</th>\n",
       "      <th>GSM650515</th>\n",
       "      <th>GSM650516</th>\n",
       "      <th>GSM650517</th>\n",
       "      <th>GSM650518</th>\n",
       "      <th>GSM650519</th>\n",
       "      <th>GSM650520</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM650644</th>\n",
       "      <th>GSM650645</th>\n",
       "      <th>GSM650646</th>\n",
       "      <th>GSM650647</th>\n",
       "      <th>GSM650648</th>\n",
       "      <th>GSM650649</th>\n",
       "      <th>GSM650651</th>\n",
       "      <th>GSM650652</th>\n",
       "      <th>GSM650653</th>\n",
       "      <th>GSM650654</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>41.632590</td>\n",
       "      <td>83.282020</td>\n",
       "      <td>52.959000</td>\n",
       "      <td>44.688750</td>\n",
       "      <td>67.330120</td>\n",
       "      <td>60.513170</td>\n",
       "      <td>44.421540</td>\n",
       "      <td>26.769420</td>\n",
       "      <td>58.695100</td>\n",
       "      <td>46.864560</td>\n",
       "      <td>...</td>\n",
       "      <td>75.627900</td>\n",
       "      <td>63.707620</td>\n",
       "      <td>33.292570</td>\n",
       "      <td>46.106280</td>\n",
       "      <td>54.652120</td>\n",
       "      <td>49.665740</td>\n",
       "      <td>64.701680</td>\n",
       "      <td>69.443120</td>\n",
       "      <td>26.259120</td>\n",
       "      <td>59.140340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>8.174400</td>\n",
       "      <td>1.181510</td>\n",
       "      <td>10.394910</td>\n",
       "      <td>9.956910</td>\n",
       "      <td>16.326440</td>\n",
       "      <td>0.197270</td>\n",
       "      <td>7.871850</td>\n",
       "      <td>0.628880</td>\n",
       "      <td>4.194580</td>\n",
       "      <td>0.004840</td>\n",
       "      <td>...</td>\n",
       "      <td>10.395270</td>\n",
       "      <td>0.241240</td>\n",
       "      <td>15.681010</td>\n",
       "      <td>5.025300</td>\n",
       "      <td>7.780800</td>\n",
       "      <td>8.694080</td>\n",
       "      <td>1.798410</td>\n",
       "      <td>0.017850</td>\n",
       "      <td>17.689140</td>\n",
       "      <td>5.513350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>95.640040</td>\n",
       "      <td>37.203130</td>\n",
       "      <td>69.185405</td>\n",
       "      <td>42.768305</td>\n",
       "      <td>46.413910</td>\n",
       "      <td>60.622755</td>\n",
       "      <td>59.372435</td>\n",
       "      <td>39.065420</td>\n",
       "      <td>26.172005</td>\n",
       "      <td>51.082100</td>\n",
       "      <td>...</td>\n",
       "      <td>30.916605</td>\n",
       "      <td>48.519265</td>\n",
       "      <td>46.315170</td>\n",
       "      <td>95.476055</td>\n",
       "      <td>52.441430</td>\n",
       "      <td>73.571020</td>\n",
       "      <td>52.925425</td>\n",
       "      <td>36.266050</td>\n",
       "      <td>106.746100</td>\n",
       "      <td>63.756045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>137.957640</td>\n",
       "      <td>81.007140</td>\n",
       "      <td>142.215430</td>\n",
       "      <td>62.315015</td>\n",
       "      <td>90.522845</td>\n",
       "      <td>96.440085</td>\n",
       "      <td>88.301515</td>\n",
       "      <td>54.964955</td>\n",
       "      <td>66.736025</td>\n",
       "      <td>55.187155</td>\n",
       "      <td>...</td>\n",
       "      <td>74.264400</td>\n",
       "      <td>124.375640</td>\n",
       "      <td>72.304515</td>\n",
       "      <td>171.866215</td>\n",
       "      <td>117.637435</td>\n",
       "      <td>111.514715</td>\n",
       "      <td>139.480970</td>\n",
       "      <td>79.037120</td>\n",
       "      <td>170.453820</td>\n",
       "      <td>139.043470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>9.691710</td>\n",
       "      <td>25.732460</td>\n",
       "      <td>32.337940</td>\n",
       "      <td>31.629130</td>\n",
       "      <td>77.091910</td>\n",
       "      <td>26.903060</td>\n",
       "      <td>34.035480</td>\n",
       "      <td>140.055140</td>\n",
       "      <td>177.304150</td>\n",
       "      <td>118.263660</td>\n",
       "      <td>...</td>\n",
       "      <td>12.995710</td>\n",
       "      <td>34.047090</td>\n",
       "      <td>30.125770</td>\n",
       "      <td>98.480480</td>\n",
       "      <td>252.059920</td>\n",
       "      <td>32.989740</td>\n",
       "      <td>58.633670</td>\n",
       "      <td>95.902850</td>\n",
       "      <td>129.589230</td>\n",
       "      <td>41.255640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>9.732460</td>\n",
       "      <td>1.848590</td>\n",
       "      <td>7.718090</td>\n",
       "      <td>12.207260</td>\n",
       "      <td>8.464150</td>\n",
       "      <td>7.863050</td>\n",
       "      <td>6.603870</td>\n",
       "      <td>9.216050</td>\n",
       "      <td>7.730030</td>\n",
       "      <td>11.694620</td>\n",
       "      <td>...</td>\n",
       "      <td>5.714060</td>\n",
       "      <td>7.159580</td>\n",
       "      <td>5.538420</td>\n",
       "      <td>18.443030</td>\n",
       "      <td>6.041980</td>\n",
       "      <td>11.726640</td>\n",
       "      <td>9.879120</td>\n",
       "      <td>12.435280</td>\n",
       "      <td>6.669430</td>\n",
       "      <td>10.415620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>90.540157</td>\n",
       "      <td>305.102487</td>\n",
       "      <td>202.538620</td>\n",
       "      <td>342.711103</td>\n",
       "      <td>309.137360</td>\n",
       "      <td>230.600267</td>\n",
       "      <td>312.511813</td>\n",
       "      <td>314.228910</td>\n",
       "      <td>377.777233</td>\n",
       "      <td>259.726963</td>\n",
       "      <td>...</td>\n",
       "      <td>381.703610</td>\n",
       "      <td>302.457647</td>\n",
       "      <td>263.350383</td>\n",
       "      <td>205.547687</td>\n",
       "      <td>345.772560</td>\n",
       "      <td>157.514923</td>\n",
       "      <td>245.048360</td>\n",
       "      <td>386.281097</td>\n",
       "      <td>195.116173</td>\n",
       "      <td>225.857520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>798.485975</td>\n",
       "      <td>1022.778865</td>\n",
       "      <td>707.486985</td>\n",
       "      <td>856.898565</td>\n",
       "      <td>759.016425</td>\n",
       "      <td>780.765135</td>\n",
       "      <td>939.762930</td>\n",
       "      <td>633.710205</td>\n",
       "      <td>974.900610</td>\n",
       "      <td>760.832645</td>\n",
       "      <td>...</td>\n",
       "      <td>949.547505</td>\n",
       "      <td>700.315820</td>\n",
       "      <td>1044.017740</td>\n",
       "      <td>495.855975</td>\n",
       "      <td>1123.309535</td>\n",
       "      <td>867.789670</td>\n",
       "      <td>748.405040</td>\n",
       "      <td>575.659335</td>\n",
       "      <td>957.902505</td>\n",
       "      <td>767.074170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>271.942207</td>\n",
       "      <td>254.717260</td>\n",
       "      <td>341.249793</td>\n",
       "      <td>268.922863</td>\n",
       "      <td>320.187483</td>\n",
       "      <td>304.505493</td>\n",
       "      <td>289.604057</td>\n",
       "      <td>305.160403</td>\n",
       "      <td>280.125727</td>\n",
       "      <td>288.047563</td>\n",
       "      <td>...</td>\n",
       "      <td>270.668230</td>\n",
       "      <td>322.724987</td>\n",
       "      <td>305.778420</td>\n",
       "      <td>273.458767</td>\n",
       "      <td>299.192847</td>\n",
       "      <td>222.870827</td>\n",
       "      <td>280.068597</td>\n",
       "      <td>301.974207</td>\n",
       "      <td>290.095660</td>\n",
       "      <td>254.628487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>31.204940</td>\n",
       "      <td>158.735530</td>\n",
       "      <td>47.567460</td>\n",
       "      <td>130.310685</td>\n",
       "      <td>110.856830</td>\n",
       "      <td>83.165085</td>\n",
       "      <td>91.090515</td>\n",
       "      <td>108.470675</td>\n",
       "      <td>146.081040</td>\n",
       "      <td>114.540680</td>\n",
       "      <td>...</td>\n",
       "      <td>151.940360</td>\n",
       "      <td>139.805945</td>\n",
       "      <td>87.837190</td>\n",
       "      <td>74.204715</td>\n",
       "      <td>103.908285</td>\n",
       "      <td>57.563500</td>\n",
       "      <td>103.637035</td>\n",
       "      <td>166.960215</td>\n",
       "      <td>16.300080</td>\n",
       "      <td>70.923315</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows Ã— 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM650510    GSM650512   GSM650513   GSM650514   GSM650515  \\\n",
       "A1BG       41.632590    83.282020   52.959000   44.688750   67.330120   \n",
       "A1BG-AS1    8.174400     1.181510   10.394910    9.956910   16.326440   \n",
       "A1CF       95.640040    37.203130   69.185405   42.768305   46.413910   \n",
       "A2M       137.957640    81.007140  142.215430   62.315015   90.522845   \n",
       "A2M-AS1     9.691710    25.732460   32.337940   31.629130   77.091910   \n",
       "...              ...          ...         ...         ...         ...   \n",
       "ZYG11A      9.732460     1.848590    7.718090   12.207260    8.464150   \n",
       "ZYG11B     90.540157   305.102487  202.538620  342.711103  309.137360   \n",
       "ZYX       798.485975  1022.778865  707.486985  856.898565  759.016425   \n",
       "ZZEF1     271.942207   254.717260  341.249793  268.922863  320.187483   \n",
       "ZZZ3       31.204940   158.735530   47.567460  130.310685  110.856830   \n",
       "\n",
       "           GSM650516   GSM650517   GSM650518   GSM650519   GSM650520  ...  \\\n",
       "A1BG       60.513170   44.421540   26.769420   58.695100   46.864560  ...   \n",
       "A1BG-AS1    0.197270    7.871850    0.628880    4.194580    0.004840  ...   \n",
       "A1CF       60.622755   59.372435   39.065420   26.172005   51.082100  ...   \n",
       "A2M        96.440085   88.301515   54.964955   66.736025   55.187155  ...   \n",
       "A2M-AS1    26.903060   34.035480  140.055140  177.304150  118.263660  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A      7.863050    6.603870    9.216050    7.730030   11.694620  ...   \n",
       "ZYG11B    230.600267  312.511813  314.228910  377.777233  259.726963  ...   \n",
       "ZYX       780.765135  939.762930  633.710205  974.900610  760.832645  ...   \n",
       "ZZEF1     304.505493  289.604057  305.160403  280.125727  288.047563  ...   \n",
       "ZZZ3       83.165085   91.090515  108.470675  146.081040  114.540680  ...   \n",
       "\n",
       "           GSM650644   GSM650645    GSM650646   GSM650647    GSM650648  \\\n",
       "A1BG       75.627900   63.707620    33.292570   46.106280    54.652120   \n",
       "A1BG-AS1   10.395270    0.241240    15.681010    5.025300     7.780800   \n",
       "A1CF       30.916605   48.519265    46.315170   95.476055    52.441430   \n",
       "A2M        74.264400  124.375640    72.304515  171.866215   117.637435   \n",
       "A2M-AS1    12.995710   34.047090    30.125770   98.480480   252.059920   \n",
       "...              ...         ...          ...         ...          ...   \n",
       "ZYG11A      5.714060    7.159580     5.538420   18.443030     6.041980   \n",
       "ZYG11B    381.703610  302.457647   263.350383  205.547687   345.772560   \n",
       "ZYX       949.547505  700.315820  1044.017740  495.855975  1123.309535   \n",
       "ZZEF1     270.668230  322.724987   305.778420  273.458767   299.192847   \n",
       "ZZZ3      151.940360  139.805945    87.837190   74.204715   103.908285   \n",
       "\n",
       "           GSM650649   GSM650651   GSM650652   GSM650653   GSM650654  \n",
       "A1BG       49.665740   64.701680   69.443120   26.259120   59.140340  \n",
       "A1BG-AS1    8.694080    1.798410    0.017850   17.689140    5.513350  \n",
       "A1CF       73.571020   52.925425   36.266050  106.746100   63.756045  \n",
       "A2M       111.514715  139.480970   79.037120  170.453820  139.043470  \n",
       "A2M-AS1    32.989740   58.633670   95.902850  129.589230   41.255640  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "ZYG11A     11.726640    9.879120   12.435280    6.669430   10.415620  \n",
       "ZYG11B    157.514923  245.048360  386.281097  195.116173  225.857520  \n",
       "ZYX       867.789670  748.405040  575.659335  957.902505  767.074170  \n",
       "ZZEF1     222.870827  280.068597  301.974207  290.095660  254.628487  \n",
       "ZZZ3       57.563500  103.637035  166.960215   16.300080   70.923315  \n",
       "\n",
       "[17857 rows x 99 columns]"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM650510</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9.83</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41.63259</td>\n",
       "      <td>8.17440</td>\n",
       "      <td>95.640040</td>\n",
       "      <td>137.957640</td>\n",
       "      <td>9.69171</td>\n",
       "      <td>37.034905</td>\n",
       "      <td>40.67379</td>\n",
       "      <td>...</td>\n",
       "      <td>48.976425</td>\n",
       "      <td>14.58944</td>\n",
       "      <td>23.76398</td>\n",
       "      <td>19.403560</td>\n",
       "      <td>37.706935</td>\n",
       "      <td>9.73246</td>\n",
       "      <td>90.540157</td>\n",
       "      <td>798.485975</td>\n",
       "      <td>271.942207</td>\n",
       "      <td>31.204940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650512</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.58</td>\n",
       "      <td>0.0</td>\n",
       "      <td>83.28202</td>\n",
       "      <td>1.18151</td>\n",
       "      <td>37.203130</td>\n",
       "      <td>81.007140</td>\n",
       "      <td>25.73246</td>\n",
       "      <td>15.882900</td>\n",
       "      <td>38.60556</td>\n",
       "      <td>...</td>\n",
       "      <td>125.516175</td>\n",
       "      <td>114.41336</td>\n",
       "      <td>48.14110</td>\n",
       "      <td>60.779100</td>\n",
       "      <td>95.003493</td>\n",
       "      <td>1.84859</td>\n",
       "      <td>305.102487</td>\n",
       "      <td>1022.778865</td>\n",
       "      <td>254.717260</td>\n",
       "      <td>158.735530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650513</th>\n",
       "      <td>0.0</td>\n",
       "      <td>14.08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.95900</td>\n",
       "      <td>10.39491</td>\n",
       "      <td>69.185405</td>\n",
       "      <td>142.215430</td>\n",
       "      <td>32.33794</td>\n",
       "      <td>22.297505</td>\n",
       "      <td>28.38244</td>\n",
       "      <td>...</td>\n",
       "      <td>45.013960</td>\n",
       "      <td>22.81288</td>\n",
       "      <td>11.00663</td>\n",
       "      <td>30.196130</td>\n",
       "      <td>92.381738</td>\n",
       "      <td>7.71809</td>\n",
       "      <td>202.538620</td>\n",
       "      <td>707.486985</td>\n",
       "      <td>341.249793</td>\n",
       "      <td>47.567460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650514</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.67</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.68875</td>\n",
       "      <td>9.95691</td>\n",
       "      <td>42.768305</td>\n",
       "      <td>62.315015</td>\n",
       "      <td>31.62913</td>\n",
       "      <td>28.699715</td>\n",
       "      <td>36.18376</td>\n",
       "      <td>...</td>\n",
       "      <td>88.122150</td>\n",
       "      <td>82.61474</td>\n",
       "      <td>30.05556</td>\n",
       "      <td>58.672835</td>\n",
       "      <td>108.238012</td>\n",
       "      <td>12.20726</td>\n",
       "      <td>342.711103</td>\n",
       "      <td>856.898565</td>\n",
       "      <td>268.922863</td>\n",
       "      <td>130.310685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650515</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>67.33012</td>\n",
       "      <td>16.32644</td>\n",
       "      <td>46.413910</td>\n",
       "      <td>90.522845</td>\n",
       "      <td>77.09191</td>\n",
       "      <td>21.434480</td>\n",
       "      <td>33.22135</td>\n",
       "      <td>...</td>\n",
       "      <td>84.270565</td>\n",
       "      <td>34.40494</td>\n",
       "      <td>37.61480</td>\n",
       "      <td>56.821985</td>\n",
       "      <td>108.401883</td>\n",
       "      <td>8.46415</td>\n",
       "      <td>309.137360</td>\n",
       "      <td>759.016425</td>\n",
       "      <td>320.187483</td>\n",
       "      <td>110.856830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650649</th>\n",
       "      <td>0.0</td>\n",
       "      <td>14.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.66574</td>\n",
       "      <td>8.69408</td>\n",
       "      <td>73.571020</td>\n",
       "      <td>111.514715</td>\n",
       "      <td>32.98974</td>\n",
       "      <td>35.383350</td>\n",
       "      <td>39.55603</td>\n",
       "      <td>...</td>\n",
       "      <td>56.185930</td>\n",
       "      <td>39.00678</td>\n",
       "      <td>14.43924</td>\n",
       "      <td>16.463120</td>\n",
       "      <td>42.707600</td>\n",
       "      <td>11.72664</td>\n",
       "      <td>157.514923</td>\n",
       "      <td>867.789670</td>\n",
       "      <td>222.870827</td>\n",
       "      <td>57.563500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650651</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.70168</td>\n",
       "      <td>1.79841</td>\n",
       "      <td>52.925425</td>\n",
       "      <td>139.480970</td>\n",
       "      <td>58.63367</td>\n",
       "      <td>24.866470</td>\n",
       "      <td>42.25776</td>\n",
       "      <td>...</td>\n",
       "      <td>70.175085</td>\n",
       "      <td>45.32123</td>\n",
       "      <td>41.18217</td>\n",
       "      <td>43.689005</td>\n",
       "      <td>85.825290</td>\n",
       "      <td>9.87912</td>\n",
       "      <td>245.048360</td>\n",
       "      <td>748.405040</td>\n",
       "      <td>280.068597</td>\n",
       "      <td>103.637035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650652</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>69.44312</td>\n",
       "      <td>0.01785</td>\n",
       "      <td>36.266050</td>\n",
       "      <td>79.037120</td>\n",
       "      <td>95.90285</td>\n",
       "      <td>16.773860</td>\n",
       "      <td>29.24067</td>\n",
       "      <td>...</td>\n",
       "      <td>117.394060</td>\n",
       "      <td>74.97703</td>\n",
       "      <td>54.66800</td>\n",
       "      <td>66.129325</td>\n",
       "      <td>96.419222</td>\n",
       "      <td>12.43528</td>\n",
       "      <td>386.281097</td>\n",
       "      <td>575.659335</td>\n",
       "      <td>301.974207</td>\n",
       "      <td>166.960215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650653</th>\n",
       "      <td>0.0</td>\n",
       "      <td>15.92</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.25912</td>\n",
       "      <td>17.68914</td>\n",
       "      <td>106.746100</td>\n",
       "      <td>170.453820</td>\n",
       "      <td>129.58923</td>\n",
       "      <td>51.762890</td>\n",
       "      <td>45.00520</td>\n",
       "      <td>...</td>\n",
       "      <td>28.036055</td>\n",
       "      <td>14.36849</td>\n",
       "      <td>19.99744</td>\n",
       "      <td>26.536545</td>\n",
       "      <td>52.304638</td>\n",
       "      <td>6.66943</td>\n",
       "      <td>195.116173</td>\n",
       "      <td>957.902505</td>\n",
       "      <td>290.095660</td>\n",
       "      <td>16.300080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM650654</th>\n",
       "      <td>0.0</td>\n",
       "      <td>16.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.14034</td>\n",
       "      <td>5.51335</td>\n",
       "      <td>63.756045</td>\n",
       "      <td>139.043470</td>\n",
       "      <td>41.25564</td>\n",
       "      <td>24.304080</td>\n",
       "      <td>35.70014</td>\n",
       "      <td>...</td>\n",
       "      <td>72.817755</td>\n",
       "      <td>44.53532</td>\n",
       "      <td>37.34833</td>\n",
       "      <td>34.378430</td>\n",
       "      <td>67.996637</td>\n",
       "      <td>10.41562</td>\n",
       "      <td>225.857520</td>\n",
       "      <td>767.074170</td>\n",
       "      <td>254.628487</td>\n",
       "      <td>70.923315</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows Ã— 17860 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Anxiety disorder    Age  Gender      A1BG  A1BG-AS1        A1CF  \\\n",
       "GSM650510               1.0   9.83     0.0  41.63259   8.17440   95.640040   \n",
       "GSM650512               0.0   6.58     0.0  83.28202   1.18151   37.203130   \n",
       "GSM650513               0.0  14.08     0.0  52.95900  10.39491   69.185405   \n",
       "GSM650514               0.0   7.67     0.0  44.68875   9.95691   42.768305   \n",
       "GSM650515               0.0   6.50     0.0  67.33012  16.32644   46.413910   \n",
       "...                     ...    ...     ...       ...       ...         ...   \n",
       "GSM650649               0.0  14.00     0.0  49.66574   8.69408   73.571020   \n",
       "GSM650651               0.0  15.25     0.0  64.70168   1.79841   52.925425   \n",
       "GSM650652               0.0  15.42     0.0  69.44312   0.01785   36.266050   \n",
       "GSM650653               0.0  15.92     0.0  26.25912  17.68914  106.746100   \n",
       "GSM650654               0.0  16.00     0.0  59.14034   5.51335   63.756045   \n",
       "\n",
       "                  A2M    A2M-AS1      A2ML1     A2MP1  ...      ZWILCH  \\\n",
       "GSM650510  137.957640    9.69171  37.034905  40.67379  ...   48.976425   \n",
       "GSM650512   81.007140   25.73246  15.882900  38.60556  ...  125.516175   \n",
       "GSM650513  142.215430   32.33794  22.297505  28.38244  ...   45.013960   \n",
       "GSM650514   62.315015   31.62913  28.699715  36.18376  ...   88.122150   \n",
       "GSM650515   90.522845   77.09191  21.434480  33.22135  ...   84.270565   \n",
       "...               ...        ...        ...       ...  ...         ...   \n",
       "GSM650649  111.514715   32.98974  35.383350  39.55603  ...   56.185930   \n",
       "GSM650651  139.480970   58.63367  24.866470  42.25776  ...   70.175085   \n",
       "GSM650652   79.037120   95.90285  16.773860  29.24067  ...  117.394060   \n",
       "GSM650653  170.453820  129.58923  51.762890  45.00520  ...   28.036055   \n",
       "GSM650654  139.043470   41.25564  24.304080  35.70014  ...   72.817755   \n",
       "\n",
       "               ZWINT      ZXDA       ZXDB        ZXDC    ZYG11A      ZYG11B  \\\n",
       "GSM650510   14.58944  23.76398  19.403560   37.706935   9.73246   90.540157   \n",
       "GSM650512  114.41336  48.14110  60.779100   95.003493   1.84859  305.102487   \n",
       "GSM650513   22.81288  11.00663  30.196130   92.381738   7.71809  202.538620   \n",
       "GSM650514   82.61474  30.05556  58.672835  108.238012  12.20726  342.711103   \n",
       "GSM650515   34.40494  37.61480  56.821985  108.401883   8.46415  309.137360   \n",
       "...              ...       ...        ...         ...       ...         ...   \n",
       "GSM650649   39.00678  14.43924  16.463120   42.707600  11.72664  157.514923   \n",
       "GSM650651   45.32123  41.18217  43.689005   85.825290   9.87912  245.048360   \n",
       "GSM650652   74.97703  54.66800  66.129325   96.419222  12.43528  386.281097   \n",
       "GSM650653   14.36849  19.99744  26.536545   52.304638   6.66943  195.116173   \n",
       "GSM650654   44.53532  37.34833  34.378430   67.996637  10.41562  225.857520   \n",
       "\n",
       "                   ZYX       ZZEF1        ZZZ3  \n",
       "GSM650510   798.485975  271.942207   31.204940  \n",
       "GSM650512  1022.778865  254.717260  158.735530  \n",
       "GSM650513   707.486985  341.249793   47.567460  \n",
       "GSM650514   856.898565  268.922863  130.310685  \n",
       "GSM650515   759.016425  320.187483  110.856830  \n",
       "...                ...         ...         ...  \n",
       "GSM650649   867.789670  222.870827   57.563500  \n",
       "GSM650651   748.405040  280.068597  103.637035  \n",
       "GSM650652   575.659335  301.974207  166.960215  \n",
       "GSM650653   957.902505  290.095660   16.300080  \n",
       "GSM650654   767.074170  254.628487   70.923315  \n",
       "\n",
       "[99 rows x 17860 columns]"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 99 samples.\n",
      "For the feature 'Anxiety disorder', the least common label is '1.0' with 3 occurrences. This represents 3.03% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 4.96\n",
      "  50% (Median): 7.5\n",
      "  75%: 11.54\n",
      "Min: 2.75\n",
      "Max: 17.5\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 99 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Leukocyte gene expression variation as a function of Big 5 dimensions of human personality\"\n",
      "!Series_summary\t\"Individual differences in basal leukocyte gene expression profiles as a function of Big 5 personality dimensions\"\n",
      "!Series_overall_design\t\"Gene expression profiling was carried out on peripheral blood mononuclear cell RNA samples collected from 119 healthy adults measured for the 5 major dimensions of human personality (Neuroticism, Extraversion, Openness, Agreeableness, Conscientiousness) using the NEO-FFI 60-item personality inventory.  Personality measures are z-score standardized.  Analyses control for major demographic characteristics (age, sex, Caucasian vs Non-Caucasian race) as well as Body Mass Index (BMI), smoking (CigDay), alcohol consumption (AlcDay), and physical activity (ExerDay, hours per day).  Additional secondary analyses controlled for the presence of minor physical symptoms (MinorSymptom, e.g.,hayfever, headache), medication use (BirthControl, Antidepressant, OtherMedication), and negative affective states (NegativeAffect, standardized values of PANAS Negative Affect scale).\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Experiment type: Risk prediction\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1481100</th>\n",
       "      <th>GSM1481101</th>\n",
       "      <th>GSM1481102</th>\n",
       "      <th>GSM1481103</th>\n",
       "      <th>GSM1481104</th>\n",
       "      <th>GSM1481105</th>\n",
       "      <th>GSM1481106</th>\n",
       "      <th>GSM1481107</th>\n",
       "      <th>GSM1481108</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1481209</th>\n",
       "      <th>GSM1481210</th>\n",
       "      <th>GSM1481211</th>\n",
       "      <th>GSM1481212</th>\n",
       "      <th>GSM1481213</th>\n",
       "      <th>GSM1481214</th>\n",
       "      <th>GSM1481215</th>\n",
       "      <th>GSM1481216</th>\n",
       "      <th>GSM1481217</th>\n",
       "      <th>GSM1481218</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 21</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 21</td>\n",
       "      <td>age: 21</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 24</td>\n",
       "      <td>age: 20</td>\n",
       "      <td>age: 26</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 32</td>\n",
       "      <td>age: 27</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 26</td>\n",
       "      <td>age: 18</td>\n",
       "      <td>age: 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>bmi: 23</td>\n",
       "      <td>bmi: 30</td>\n",
       "      <td>bmi: 18</td>\n",
       "      <td>bmi: 21.6</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 16</td>\n",
       "      <td>bmi: 26</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>...</td>\n",
       "      <td>bmi: 20</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 20</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 21.9</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 23</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: missing</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 7</td>\n",
       "      <td>cigday: 12</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>alcday: 0.4</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 1.3</td>\n",
       "      <td>alcday: 0.9</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>alcday: 6</td>\n",
       "      <td>alcday: 0.2</td>\n",
       "      <td>alcday: 0.6</td>\n",
       "      <td>alcday: 1</td>\n",
       "      <td>alcday: 0.1</td>\n",
       "      <td>alcday: 0.4</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 0</td>\n",
       "      <td>alcday: 1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>exerday: 0.5</td>\n",
       "      <td>exerday: 0.9</td>\n",
       "      <td>exerday: 0.6</td>\n",
       "      <td>exerday: 0.6</td>\n",
       "      <td>exerday: 1</td>\n",
       "      <td>exerday: 0.1</td>\n",
       "      <td>exerday: 0.1</td>\n",
       "      <td>exerday: 0.3</td>\n",
       "      <td>exerday: 0.5</td>\n",
       "      <td>...</td>\n",
       "      <td>exerday: 0.1</td>\n",
       "      <td>exerday: 0.1</td>\n",
       "      <td>exerday: 0.7</td>\n",
       "      <td>exerday: 1</td>\n",
       "      <td>exerday: 1.1</td>\n",
       "      <td>exerday: 0.7</td>\n",
       "      <td>exerday: 1</td>\n",
       "      <td>exerday: 1</td>\n",
       "      <td>exerday: 1</td>\n",
       "      <td>exerday: 0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: missing</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "      <td>minorsymptom: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 1</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: missing</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "      <td>birthcontrol: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 1</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: missing</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "      <td>antidepressant: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 1</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: missing</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "      <td>othermedication: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>negativeaffect: 0.162</td>\n",
       "      <td>negativeaffect: 1.319</td>\n",
       "      <td>negativeaffect: 1.65</td>\n",
       "      <td>negativeaffect: -0.83</td>\n",
       "      <td>negativeaffect: 2.311</td>\n",
       "      <td>negativeaffect: -0.664</td>\n",
       "      <td>negativeaffect: -0.003</td>\n",
       "      <td>negativeaffect: 1.319</td>\n",
       "      <td>negativeaffect: -0.83</td>\n",
       "      <td>...</td>\n",
       "      <td>negativeaffect: 2.145</td>\n",
       "      <td>negativeaffect: -0.334</td>\n",
       "      <td>negativeaffect: -0.334</td>\n",
       "      <td>negativeaffect: -0.334</td>\n",
       "      <td>negativeaffect: -0.334</td>\n",
       "      <td>negativeaffect: -0.334</td>\n",
       "      <td>negativeaffect: -0.664</td>\n",
       "      <td>negativeaffect: -0.83</td>\n",
       "      <td>negativeaffect: -0.83</td>\n",
       "      <td>negativeaffect: 0.162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>neuroticism: 0.18</td>\n",
       "      <td>neuroticism: -0.96</td>\n",
       "      <td>neuroticism: 2.58</td>\n",
       "      <td>neuroticism: -1.47</td>\n",
       "      <td>neuroticism: 1.06</td>\n",
       "      <td>neuroticism: -0.46</td>\n",
       "      <td>neuroticism: 0.3</td>\n",
       "      <td>neuroticism: 0.94</td>\n",
       "      <td>neuroticism: 2.46</td>\n",
       "      <td>...</td>\n",
       "      <td>neuroticism: 1.32</td>\n",
       "      <td>neuroticism: 1.06</td>\n",
       "      <td>neuroticism: 0.56</td>\n",
       "      <td>neuroticism: -0.2</td>\n",
       "      <td>neuroticism: -1.09</td>\n",
       "      <td>neuroticism: 1.57</td>\n",
       "      <td>neuroticism: -0.96</td>\n",
       "      <td>neuroticism: -1.34</td>\n",
       "      <td>neuroticism: 0.56</td>\n",
       "      <td>neuroticism: 0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>extraversion: 0.46</td>\n",
       "      <td>extraversion: -0.43</td>\n",
       "      <td>extraversion: -1.01</td>\n",
       "      <td>extraversion: 0.6</td>\n",
       "      <td>extraversion: -0.72</td>\n",
       "      <td>extraversion: -0.57</td>\n",
       "      <td>extraversion: -1.16</td>\n",
       "      <td>extraversion: 0.16</td>\n",
       "      <td>extraversion: -1.75</td>\n",
       "      <td>...</td>\n",
       "      <td>extraversion: -0.57</td>\n",
       "      <td>extraversion: -0.13</td>\n",
       "      <td>extraversion: -0.13</td>\n",
       "      <td>extraversion: 0.75</td>\n",
       "      <td>extraversion: 1.19</td>\n",
       "      <td>extraversion: 0.6</td>\n",
       "      <td>extraversion: 0.75</td>\n",
       "      <td>extraversion: -0.43</td>\n",
       "      <td>extraversion: 0.75</td>\n",
       "      <td>extraversion: 0.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>openness: 0.27</td>\n",
       "      <td>openness: 0.73</td>\n",
       "      <td>openness: 2.14</td>\n",
       "      <td>openness: -0.2</td>\n",
       "      <td>openness: 0.73</td>\n",
       "      <td>openness: -1.14</td>\n",
       "      <td>openness: 0.11</td>\n",
       "      <td>openness: 0.89</td>\n",
       "      <td>openness: 1.05</td>\n",
       "      <td>...</td>\n",
       "      <td>openness: 0.73</td>\n",
       "      <td>openness: -0.98</td>\n",
       "      <td>openness: 0.11</td>\n",
       "      <td>openness: 0.11</td>\n",
       "      <td>openness: -0.83</td>\n",
       "      <td>openness: 1.52</td>\n",
       "      <td>openness: -0.2</td>\n",
       "      <td>openness: 1.52</td>\n",
       "      <td>openness: -0.98</td>\n",
       "      <td>openness: -1.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>agreeableness: -0.02</td>\n",
       "      <td>agreeableness: -0.81</td>\n",
       "      <td>agreeableness: -0.49</td>\n",
       "      <td>agreeableness: -0.02</td>\n",
       "      <td>agreeableness: -0.33</td>\n",
       "      <td>agreeableness: 0.78</td>\n",
       "      <td>agreeableness: -1.13</td>\n",
       "      <td>agreeableness: 0.14</td>\n",
       "      <td>agreeableness: 1.41</td>\n",
       "      <td>...</td>\n",
       "      <td>agreeableness: -3.19</td>\n",
       "      <td>agreeableness: -0.02</td>\n",
       "      <td>agreeableness: 0.78</td>\n",
       "      <td>agreeableness: 1.41</td>\n",
       "      <td>agreeableness: 0.78</td>\n",
       "      <td>agreeableness: 0.62</td>\n",
       "      <td>agreeableness: 0.3</td>\n",
       "      <td>agreeableness: -0.81</td>\n",
       "      <td>agreeableness: -0.97</td>\n",
       "      <td>agreeableness: 0.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>conscientiousness: -0.51</td>\n",
       "      <td>conscientiousness: 0.66</td>\n",
       "      <td>conscientiousness: 0.01</td>\n",
       "      <td>conscientiousness: -1.68</td>\n",
       "      <td>conscientiousness: 1.31</td>\n",
       "      <td>conscientiousness: -0.38</td>\n",
       "      <td>conscientiousness: 0.53</td>\n",
       "      <td>conscientiousness: -0.38</td>\n",
       "      <td>conscientiousness: -1.42</td>\n",
       "      <td>...</td>\n",
       "      <td>conscientiousness: -2.45</td>\n",
       "      <td>conscientiousness: -0.25</td>\n",
       "      <td>conscientiousness: -0.25</td>\n",
       "      <td>conscientiousness: 0.66</td>\n",
       "      <td>conscientiousness: -0.38</td>\n",
       "      <td>conscientiousness: -1.29</td>\n",
       "      <td>conscientiousness: -1.03</td>\n",
       "      <td>conscientiousness: 1.83</td>\n",
       "      <td>conscientiousness: -0.25</td>\n",
       "      <td>conscientiousness: 1.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows Ã— 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          !Sample_geo_accession                GSM1481100  \\\n",
       "0   !Sample_characteristics_ch1                   age: 21   \n",
       "1   !Sample_characteristics_ch1                   male: 0   \n",
       "2   !Sample_characteristics_ch1                   bmi: 23   \n",
       "3   !Sample_characteristics_ch1              caucasian: 1   \n",
       "4   !Sample_characteristics_ch1                 cigday: 0   \n",
       "5   !Sample_characteristics_ch1               alcday: 0.4   \n",
       "6   !Sample_characteristics_ch1              exerday: 0.5   \n",
       "7   !Sample_characteristics_ch1           minorsymptom: 0   \n",
       "8   !Sample_characteristics_ch1           birthcontrol: 0   \n",
       "9   !Sample_characteristics_ch1         antidepressant: 0   \n",
       "10  !Sample_characteristics_ch1        othermedication: 0   \n",
       "11  !Sample_characteristics_ch1     negativeaffect: 0.162   \n",
       "12  !Sample_characteristics_ch1         neuroticism: 0.18   \n",
       "13  !Sample_characteristics_ch1        extraversion: 0.46   \n",
       "14  !Sample_characteristics_ch1            openness: 0.27   \n",
       "15  !Sample_characteristics_ch1      agreeableness: -0.02   \n",
       "16  !Sample_characteristics_ch1  conscientiousness: -0.51   \n",
       "\n",
       "                 GSM1481101               GSM1481102  \\\n",
       "0                   age: 22                  age: 21   \n",
       "1                   male: 1                  male: 0   \n",
       "2                   bmi: 30                  bmi: 18   \n",
       "3              caucasian: 1             caucasian: 1   \n",
       "4                 cigday: 7               cigday: 12   \n",
       "5                 alcday: 0              alcday: 1.3   \n",
       "6              exerday: 0.9             exerday: 0.6   \n",
       "7           minorsymptom: 0          minorsymptom: 0   \n",
       "8           birthcontrol: 0          birthcontrol: 0   \n",
       "9         antidepressant: 0        antidepressant: 0   \n",
       "10       othermedication: 0       othermedication: 0   \n",
       "11    negativeaffect: 1.319     negativeaffect: 1.65   \n",
       "12       neuroticism: -0.96        neuroticism: 2.58   \n",
       "13      extraversion: -0.43      extraversion: -1.01   \n",
       "14           openness: 0.73           openness: 2.14   \n",
       "15     agreeableness: -0.81     agreeableness: -0.49   \n",
       "16  conscientiousness: 0.66  conscientiousness: 0.01   \n",
       "\n",
       "                  GSM1481103               GSM1481104  \\\n",
       "0                    age: 21                  age: 22   \n",
       "1                    male: 1                  male: 0   \n",
       "2                  bmi: 21.6                  bmi: 22   \n",
       "3               caucasian: 0             caucasian: 0   \n",
       "4                  cigday: 0                cigday: 0   \n",
       "5                alcday: 0.9                alcday: 0   \n",
       "6               exerday: 0.6               exerday: 1   \n",
       "7            minorsymptom: 0          minorsymptom: 0   \n",
       "8            birthcontrol: 0          birthcontrol: 0   \n",
       "9          antidepressant: 0        antidepressant: 0   \n",
       "10        othermedication: 0       othermedication: 0   \n",
       "11     negativeaffect: -0.83    negativeaffect: 2.311   \n",
       "12        neuroticism: -1.47        neuroticism: 1.06   \n",
       "13         extraversion: 0.6      extraversion: -0.72   \n",
       "14            openness: -0.2           openness: 0.73   \n",
       "15      agreeableness: -0.02     agreeableness: -0.33   \n",
       "16  conscientiousness: -1.68  conscientiousness: 1.31   \n",
       "\n",
       "                  GSM1481105               GSM1481106  \\\n",
       "0                    age: 23                  age: 23   \n",
       "1                    male: 0                  male: 0   \n",
       "2                    bmi: 22                  bmi: 16   \n",
       "3               caucasian: 0             caucasian: 0   \n",
       "4                  cigday: 0                cigday: 0   \n",
       "5                  alcday: 0                alcday: 0   \n",
       "6               exerday: 0.1             exerday: 0.1   \n",
       "7            minorsymptom: 0          minorsymptom: 0   \n",
       "8            birthcontrol: 0          birthcontrol: 0   \n",
       "9          antidepressant: 0        antidepressant: 0   \n",
       "10        othermedication: 0       othermedication: 0   \n",
       "11    negativeaffect: -0.664   negativeaffect: -0.003   \n",
       "12        neuroticism: -0.46         neuroticism: 0.3   \n",
       "13       extraversion: -0.57      extraversion: -1.16   \n",
       "14           openness: -1.14           openness: 0.11   \n",
       "15       agreeableness: 0.78     agreeableness: -1.13   \n",
       "16  conscientiousness: -0.38  conscientiousness: 0.53   \n",
       "\n",
       "                  GSM1481107                GSM1481108  ...  \\\n",
       "0                    age: 23                   age: 23  ...   \n",
       "1                    male: 0                   male: 0  ...   \n",
       "2                    bmi: 26                   bmi: 24  ...   \n",
       "3               caucasian: 0              caucasian: 0  ...   \n",
       "4                  cigday: 0                 cigday: 0  ...   \n",
       "5                  alcday: 0                 alcday: 0  ...   \n",
       "6               exerday: 0.3              exerday: 0.5  ...   \n",
       "7            minorsymptom: 0           minorsymptom: 0  ...   \n",
       "8            birthcontrol: 0           birthcontrol: 0  ...   \n",
       "9          antidepressant: 1         antidepressant: 0  ...   \n",
       "10        othermedication: 0        othermedication: 0  ...   \n",
       "11     negativeaffect: 1.319     negativeaffect: -0.83  ...   \n",
       "12         neuroticism: 0.94         neuroticism: 2.46  ...   \n",
       "13        extraversion: 0.16       extraversion: -1.75  ...   \n",
       "14            openness: 0.89            openness: 1.05  ...   \n",
       "15       agreeableness: 0.14       agreeableness: 1.41  ...   \n",
       "16  conscientiousness: -0.38  conscientiousness: -1.42  ...   \n",
       "\n",
       "                  GSM1481209                GSM1481210  \\\n",
       "0                    age: 24                   age: 20   \n",
       "1                    male: 1                   male: 1   \n",
       "2                    bmi: 20                   bmi: 24   \n",
       "3               caucasian: 0              caucasian: 1   \n",
       "4                  cigday: 0                 cigday: 0   \n",
       "5                  alcday: 6               alcday: 0.2   \n",
       "6               exerday: 0.1              exerday: 0.1   \n",
       "7            minorsymptom: 0           minorsymptom: 0   \n",
       "8            birthcontrol: 0           birthcontrol: 0   \n",
       "9          antidepressant: 0         antidepressant: 0   \n",
       "10        othermedication: 0        othermedication: 0   \n",
       "11     negativeaffect: 2.145    negativeaffect: -0.334   \n",
       "12         neuroticism: 1.32         neuroticism: 1.06   \n",
       "13       extraversion: -0.57       extraversion: -0.13   \n",
       "14            openness: 0.73           openness: -0.98   \n",
       "15      agreeableness: -3.19      agreeableness: -0.02   \n",
       "16  conscientiousness: -2.45  conscientiousness: -0.25   \n",
       "\n",
       "                  GSM1481211               GSM1481212  \\\n",
       "0                    age: 26                  age: 22   \n",
       "1                    male: 0                  male: 0   \n",
       "2                    bmi: 20                  bmi: 22   \n",
       "3               caucasian: 1             caucasian: 1   \n",
       "4                  cigday: 0                cigday: 0   \n",
       "5                alcday: 0.6                alcday: 1   \n",
       "6               exerday: 0.7               exerday: 1   \n",
       "7            minorsymptom: 0          minorsymptom: 0   \n",
       "8            birthcontrol: 1          birthcontrol: 0   \n",
       "9          antidepressant: 0        antidepressant: 0   \n",
       "10        othermedication: 0       othermedication: 1   \n",
       "11    negativeaffect: -0.334   negativeaffect: -0.334   \n",
       "12         neuroticism: 0.56        neuroticism: -0.2   \n",
       "13       extraversion: -0.13       extraversion: 0.75   \n",
       "14            openness: 0.11           openness: 0.11   \n",
       "15       agreeableness: 0.78      agreeableness: 1.41   \n",
       "16  conscientiousness: -0.25  conscientiousness: 0.66   \n",
       "\n",
       "                  GSM1481213                GSM1481214  \\\n",
       "0                    age: 32                   age: 27   \n",
       "1                    male: 1                   male: 1   \n",
       "2                  bmi: 21.9                   bmi: 22   \n",
       "3               caucasian: 0        caucasian: missing   \n",
       "4                  cigday: 0                 cigday: 0   \n",
       "5                alcday: 0.1               alcday: 0.4   \n",
       "6               exerday: 1.1              exerday: 0.7   \n",
       "7            minorsymptom: 0     minorsymptom: missing   \n",
       "8            birthcontrol: 0     birthcontrol: missing   \n",
       "9          antidepressant: 0   antidepressant: missing   \n",
       "10        othermedication: 0  othermedication: missing   \n",
       "11    negativeaffect: -0.334    negativeaffect: -0.334   \n",
       "12        neuroticism: -1.09         neuroticism: 1.57   \n",
       "13        extraversion: 1.19         extraversion: 0.6   \n",
       "14           openness: -0.83            openness: 1.52   \n",
       "15       agreeableness: 0.78       agreeableness: 0.62   \n",
       "16  conscientiousness: -0.38  conscientiousness: -1.29   \n",
       "\n",
       "                  GSM1481215               GSM1481216  \\\n",
       "0                    age: 22                  age: 26   \n",
       "1                    male: 1                  male: 1   \n",
       "2                    bmi: 24                  bmi: 23   \n",
       "3               caucasian: 1             caucasian: 1   \n",
       "4                  cigday: 0                cigday: 0   \n",
       "5                  alcday: 0                alcday: 0   \n",
       "6                 exerday: 1               exerday: 1   \n",
       "7            minorsymptom: 0          minorsymptom: 0   \n",
       "8            birthcontrol: 0          birthcontrol: 0   \n",
       "9          antidepressant: 0        antidepressant: 0   \n",
       "10        othermedication: 0       othermedication: 0   \n",
       "11    negativeaffect: -0.664    negativeaffect: -0.83   \n",
       "12        neuroticism: -0.96       neuroticism: -1.34   \n",
       "13        extraversion: 0.75      extraversion: -0.43   \n",
       "14            openness: -0.2           openness: 1.52   \n",
       "15        agreeableness: 0.3     agreeableness: -0.81   \n",
       "16  conscientiousness: -1.03  conscientiousness: 1.83   \n",
       "\n",
       "                  GSM1481217               GSM1481218  \n",
       "0                    age: 18                  age: 20  \n",
       "1                    male: 1                  male: 0  \n",
       "2                    bmi: 24                  bmi: 36  \n",
       "3               caucasian: 1             caucasian: 1  \n",
       "4                  cigday: 0                cigday: 0  \n",
       "5                  alcday: 0              alcday: 1.3  \n",
       "6                 exerday: 1             exerday: 0.4  \n",
       "7            minorsymptom: 0          minorsymptom: 0  \n",
       "8            birthcontrol: 0          birthcontrol: 0  \n",
       "9          antidepressant: 0        antidepressant: 0  \n",
       "10        othermedication: 0       othermedication: 0  \n",
       "11     negativeaffect: -0.83    negativeaffect: 0.162  \n",
       "12         neuroticism: 0.56        neuroticism: 0.05  \n",
       "13        extraversion: 0.75       extraversion: 0.46  \n",
       "14           openness: -0.98          openness: -1.14  \n",
       "15      agreeableness: -0.97      agreeableness: 0.62  \n",
       "16  conscientiousness: -0.25  conscientiousness: 1.05  \n",
       "\n",
       "[17 rows x 120 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No trait convert\n",
    "cohort = accession_num = \"GSE60491\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'bmi: 23', 'bmi: 30', 'bmi: 18',\n",
       "       'bmi: 21.6', 'bmi: 22', 'bmi: 16', 'bmi: 26', 'bmi: 24', 'bmi: 31',\n",
       "       'bmi: 21', 'bmi: missing', 'bmi: 19', 'bmi: 25', 'bmi: 17.5',\n",
       "       'bmi: 20', 'bmi: 28', 'bmi: 29', 'bmi: 35', 'bmi: 17', 'bmi: 27',\n",
       "       'bmi: 21.9', 'bmi: 36'], dtype=object)"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# è¿™ä¸ªå‡½æ•°å°†ç»„ç»‡ç±»åž‹ï¼ˆtissue typeï¼‰è½¬æ¢ä¸ºæœ‰å…³ç™«ç—«å­˜åœ¨ä¸Žå¦çš„äºŒè¿›åˆ¶å€¼ã€‚\n",
    "# å®ƒæ˜¯åŸºäºŽç‰¹å®šçš„å‡è®¾ï¼Œå³å¦‚æžœç»„ç»‡ç±»åž‹æ˜¯â€œèƒ°è…ºå¯¼ç®¡è…ºç™Œâ€ï¼ˆPancreatic Ductal Adenocarcinomaï¼‰ï¼Œåˆ™è®¤ä¸ºç™«ç—«å­˜åœ¨ï¼ˆè¿”å›ž1ï¼‰ï¼›å¦åˆ™ï¼Œè®¤ä¸ºç™«ç—«ä¸å­˜åœ¨ï¼ˆè¿”å›ž0ï¼‰ã€‚\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'condition: tumor':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Leukocyte gene expression variation as a function of Big 5 dimensions of human personality\"\n",
      "!Series_summary\t\"Individual differences in basal leukocyte gene expression profiles as a function of Big 5 personality dimensions\"\n",
      "!Series_overall_design\t\"Gene expression profiling was carried out on peripheral blood mononuclear cell RNA samples collected from 119 healthy adults measured for the 5 major dimensions of human personality (Neuroticism, Extraversion, Openness, Agreeableness, Conscientiousness) using the NEO-FFI 60-item personality inventory.  Personality measures are z-score standardized.  Analyses control for major demographic characteristics (age, sex, Caucasian vs Non-Caucasian race) as well as Body Mass Index (BMI), smoking (CigDay), alcohol consumption (AlcDay), and physical activity (ExerDay, hours per day).  Additional secondary analyses controlled for the presence of minor physical symptoms (MinorSymptom, e.g.,hayfever, headache), medication use (BirthControl, Antidepressant, OtherMedication), and negative affective states (NegativeAffect, standardized values of PANAS Negative Affect scale).\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Experiment type: Risk prediction\"\n"
     ]
    }
   ],
   "source": [
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1481100</th>\n",
       "      <th>GSM1481101</th>\n",
       "      <th>GSM1481102</th>\n",
       "      <th>GSM1481103</th>\n",
       "      <th>GSM1481104</th>\n",
       "      <th>GSM1481105</th>\n",
       "      <th>GSM1481106</th>\n",
       "      <th>GSM1481107</th>\n",
       "      <th>GSM1481108</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1481209</th>\n",
       "      <th>GSM1481210</th>\n",
       "      <th>GSM1481211</th>\n",
       "      <th>GSM1481212</th>\n",
       "      <th>GSM1481213</th>\n",
       "      <th>GSM1481214</th>\n",
       "      <th>GSM1481215</th>\n",
       "      <th>GSM1481216</th>\n",
       "      <th>GSM1481217</th>\n",
       "      <th>GSM1481218</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 21</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 21</td>\n",
       "      <td>age: 21</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 24</td>\n",
       "      <td>age: 20</td>\n",
       "      <td>age: 26</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 32</td>\n",
       "      <td>age: 27</td>\n",
       "      <td>age: 22</td>\n",
       "      <td>age: 26</td>\n",
       "      <td>age: 18</td>\n",
       "      <td>age: 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 0</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 1</td>\n",
       "      <td>male: 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>bmi: 23</td>\n",
       "      <td>bmi: 30</td>\n",
       "      <td>bmi: 18</td>\n",
       "      <td>bmi: 21.6</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 16</td>\n",
       "      <td>bmi: 26</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>...</td>\n",
       "      <td>bmi: 20</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 20</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 21.9</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 23</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 0</td>\n",
       "      <td>caucasian: missing</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "      <td>caucasian: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 7</td>\n",
       "      <td>cigday: 12</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "      <td>cigday: 0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 120 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession    GSM1481100    GSM1481101    GSM1481102  \\\n",
       "0  !Sample_characteristics_ch1       age: 21       age: 22       age: 21   \n",
       "1  !Sample_characteristics_ch1       male: 0       male: 1       male: 0   \n",
       "2  !Sample_characteristics_ch1       bmi: 23       bmi: 30       bmi: 18   \n",
       "3  !Sample_characteristics_ch1  caucasian: 1  caucasian: 1  caucasian: 1   \n",
       "4  !Sample_characteristics_ch1     cigday: 0     cigday: 7    cigday: 12   \n",
       "\n",
       "     GSM1481103    GSM1481104    GSM1481105    GSM1481106    GSM1481107  \\\n",
       "0       age: 21       age: 22       age: 23       age: 23       age: 23   \n",
       "1       male: 1       male: 0       male: 0       male: 0       male: 0   \n",
       "2     bmi: 21.6       bmi: 22       bmi: 22       bmi: 16       bmi: 26   \n",
       "3  caucasian: 0  caucasian: 0  caucasian: 0  caucasian: 0  caucasian: 0   \n",
       "4     cigday: 0     cigday: 0     cigday: 0     cigday: 0     cigday: 0   \n",
       "\n",
       "     GSM1481108  ...    GSM1481209    GSM1481210    GSM1481211    GSM1481212  \\\n",
       "0       age: 23  ...       age: 24       age: 20       age: 26       age: 22   \n",
       "1       male: 0  ...       male: 1       male: 1       male: 0       male: 0   \n",
       "2       bmi: 24  ...       bmi: 20       bmi: 24       bmi: 20       bmi: 22   \n",
       "3  caucasian: 0  ...  caucasian: 0  caucasian: 1  caucasian: 1  caucasian: 1   \n",
       "4     cigday: 0  ...     cigday: 0     cigday: 0     cigday: 0     cigday: 0   \n",
       "\n",
       "     GSM1481213          GSM1481214    GSM1481215    GSM1481216    GSM1481217  \\\n",
       "0       age: 32             age: 27       age: 22       age: 26       age: 18   \n",
       "1       male: 1             male: 1       male: 1       male: 1       male: 1   \n",
       "2     bmi: 21.9             bmi: 22       bmi: 24       bmi: 23       bmi: 24   \n",
       "3  caucasian: 0  caucasian: missing  caucasian: 1  caucasian: 1  caucasian: 1   \n",
       "4     cigday: 0           cigday: 0     cigday: 0     cigday: 0     cigday: 0   \n",
       "\n",
       "     GSM1481218  \n",
       "0       age: 20  \n",
       "1       male: 0  \n",
       "2       bmi: 36  \n",
       "3  caucasian: 1  \n",
       "4     cigday: 0  \n",
       "\n",
       "[5 rows x 120 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['age: 21',\n",
       "  'age: 22',\n",
       "  'age: 23',\n",
       "  'age: 33',\n",
       "  'age: 20',\n",
       "  'age: 34',\n",
       "  'age: 19',\n",
       "  'age: 27',\n",
       "  'age: 53',\n",
       "  'age: 25',\n",
       "  'age: 26',\n",
       "  'age: 45',\n",
       "  'age: 38',\n",
       "  'age: 29',\n",
       "  'age: 30',\n",
       "  'age: 28',\n",
       "  'age: 18',\n",
       "  'age: 24',\n",
       "  'age: 59',\n",
       "  'age: 35',\n",
       "  'age: 51',\n",
       "  'age: 50',\n",
       "  'age: 32'],\n",
       " 1: ['male: 0', 'male: 1'],\n",
       " 2: ['bmi: 23',\n",
       "  'bmi: 30',\n",
       "  'bmi: 18',\n",
       "  'bmi: 21.6',\n",
       "  'bmi: 22',\n",
       "  'bmi: 16',\n",
       "  'bmi: 26',\n",
       "  'bmi: 24',\n",
       "  'bmi: 31',\n",
       "  'bmi: 21',\n",
       "  'bmi: missing',\n",
       "  'bmi: 19',\n",
       "  'bmi: 25',\n",
       "  'bmi: 17.5',\n",
       "  'bmi: 20',\n",
       "  'bmi: 28',\n",
       "  'bmi: 29',\n",
       "  'bmi: 35',\n",
       "  'bmi: 17',\n",
       "  'bmi: 27',\n",
       "  'bmi: 21.9',\n",
       "  'bmi: 36'],\n",
       " 3: ['caucasian: 1', 'caucasian: 0', 'caucasian: missing'],\n",
       " 4: ['cigday: 0',\n",
       "  'cigday: 7',\n",
       "  'cigday: 12',\n",
       "  'cigday: 3',\n",
       "  'cigday: 5',\n",
       "  'cigday: 0.1',\n",
       "  'cigday: 1.4',\n",
       "  'cigday: 20'],\n",
       " 5: ['alcday: 0.4',\n",
       "  'alcday: 0',\n",
       "  'alcday: 1.3',\n",
       "  'alcday: 0.9',\n",
       "  'alcday: 2',\n",
       "  'alcday: 1',\n",
       "  'alcday: 0.1',\n",
       "  'alcday: 1.4',\n",
       "  'alcday: 0.6',\n",
       "  'alcday: 2.9',\n",
       "  'alcday: 0.3',\n",
       "  'alcday: 1.7',\n",
       "  'alcday: 3',\n",
       "  'alcday: 1.1',\n",
       "  'alcday: 0.7',\n",
       "  'alcday: 2.3',\n",
       "  'alcday: 1.5',\n",
       "  'alcday: 2.2',\n",
       "  'alcday: 6',\n",
       "  'alcday: 0.2'],\n",
       " 6: ['exerday: 0.5',\n",
       "  'exerday: 0.9',\n",
       "  'exerday: 0.6',\n",
       "  'exerday: 1',\n",
       "  'exerday: 0.1',\n",
       "  'exerday: 0.3',\n",
       "  'exerday: 0',\n",
       "  'exerday: 2',\n",
       "  'exerday: 1.5',\n",
       "  'exerday: 0.7',\n",
       "  'exerday: 0.4',\n",
       "  'exerday: 0.8',\n",
       "  'exerday: 0.75',\n",
       "  'exerday: 1.4',\n",
       "  'exerday: 3',\n",
       "  'exerday: 1.1'],\n",
       " 7: ['minorsymptom: 0', 'minorsymptom: missing', 'minorsymptom: 1'],\n",
       " 8: ['birthcontrol: 0', 'birthcontrol: 1', 'birthcontrol: missing'],\n",
       " 9: ['antidepressant: 0', 'antidepressant: 1', 'antidepressant: missing'],\n",
       " 10: ['othermedication: 0', 'othermedication: missing', 'othermedication: 1'],\n",
       " 11: ['negativeaffect: 0.162',\n",
       "  'negativeaffect: 1.319',\n",
       "  'negativeaffect: 1.65',\n",
       "  'negativeaffect: -0.83',\n",
       "  'negativeaffect: 2.311',\n",
       "  'negativeaffect: -0.664',\n",
       "  'negativeaffect: -0.003',\n",
       "  'negativeaffect: 2.972',\n",
       "  'negativeaffect: 4.129',\n",
       "  'negativeaffect: -0.334',\n",
       "  'negativeaffect: -0.169',\n",
       "  'negativeaffect: 0.658',\n",
       "  'negativeaffect: 1.484',\n",
       "  'negativeaffect: 0.823',\n",
       "  'negativeaffect: 1.815',\n",
       "  'negativeaffect: 2.476',\n",
       "  'negativeaffect: 0.493',\n",
       "  'negativeaffect: 0.988',\n",
       "  'negativeaffect: 2.145'],\n",
       " 12: ['neuroticism: 0.18',\n",
       "  'neuroticism: -0.96',\n",
       "  'neuroticism: 2.58',\n",
       "  'neuroticism: -1.47',\n",
       "  'neuroticism: 1.06',\n",
       "  'neuroticism: -0.46',\n",
       "  'neuroticism: 0.3',\n",
       "  'neuroticism: 0.94',\n",
       "  'neuroticism: 2.46',\n",
       "  'neuroticism: 0.81',\n",
       "  'neuroticism: 0.43',\n",
       "  'neuroticism: -1.09',\n",
       "  'neuroticism: -0.08',\n",
       "  'neuroticism: -0.71',\n",
       "  'neuroticism: -0.2',\n",
       "  'neuroticism: -1.34',\n",
       "  'neuroticism: -0.84',\n",
       "  'neuroticism: 1.19',\n",
       "  'neuroticism: 0.56',\n",
       "  'neuroticism: 0.05',\n",
       "  'neuroticism: -0.33',\n",
       "  'neuroticism: 1.95',\n",
       "  'neuroticism: 2.08',\n",
       "  'neuroticism: -0.58',\n",
       "  'neuroticism: 0.68',\n",
       "  'neuroticism: -1.97',\n",
       "  'neuroticism: 1.44',\n",
       "  'neuroticism: -1.85',\n",
       "  'neuroticism: -1.22',\n",
       "  'neuroticism: 1.32'],\n",
       " 13: ['extraversion: 0.46',\n",
       "  'extraversion: -0.43',\n",
       "  'extraversion: -1.01',\n",
       "  'extraversion: 0.6',\n",
       "  'extraversion: -0.72',\n",
       "  'extraversion: -0.57',\n",
       "  'extraversion: -1.16',\n",
       "  'extraversion: 0.16',\n",
       "  'extraversion: -1.75',\n",
       "  'extraversion: 0.01',\n",
       "  'extraversion: 0.9',\n",
       "  'extraversion: 1.04',\n",
       "  'extraversion: -1.31',\n",
       "  'extraversion: 2.22',\n",
       "  'extraversion: -1.6',\n",
       "  'extraversion: 1.19',\n",
       "  'extraversion: -2.34',\n",
       "  'extraversion: -2.19',\n",
       "  'extraversion: 1.34',\n",
       "  'extraversion: 0.31',\n",
       "  'extraversion: -0.28',\n",
       "  'extraversion: -1.46',\n",
       "  'extraversion: -0.87',\n",
       "  'extraversion: -0.13',\n",
       "  'extraversion: 1.93',\n",
       "  'extraversion: 1.63',\n",
       "  'extraversion: 2.07',\n",
       "  'extraversion: 0.75',\n",
       "  'extraversion: 1.49'],\n",
       " 14: ['openness: 0.27',\n",
       "  'openness: 0.73',\n",
       "  'openness: 2.14',\n",
       "  'openness: -0.2',\n",
       "  'openness: -1.14',\n",
       "  'openness: 0.11',\n",
       "  'openness: 0.89',\n",
       "  'openness: 1.05',\n",
       "  'openness: -0.36',\n",
       "  'openness: -0.98',\n",
       "  'openness: 1.67',\n",
       "  'openness: -1.92',\n",
       "  'openness: -0.83',\n",
       "  'openness: -2.39',\n",
       "  'openness: 1.52',\n",
       "  'openness: -0.67',\n",
       "  'openness: -0.05',\n",
       "  'openness: 0.42',\n",
       "  'openness: -0.52',\n",
       "  'openness: 1.2',\n",
       "  'openness: -1.77',\n",
       "  'openness: 0.58',\n",
       "  'openness: 2.45',\n",
       "  'openness: -1.3',\n",
       "  'openness: 1.36',\n",
       "  'openness: 1.83',\n",
       "  'openness: -1.61',\n",
       "  'openness: -2.23',\n",
       "  'openness: -2.08',\n",
       "  'openness: -2.7'],\n",
       " 15: ['agreeableness: -0.02',\n",
       "  'agreeableness: -0.81',\n",
       "  'agreeableness: -0.49',\n",
       "  'agreeableness: -0.33',\n",
       "  'agreeableness: 0.78',\n",
       "  'agreeableness: -1.13',\n",
       "  'agreeableness: 0.14',\n",
       "  'agreeableness: 1.41',\n",
       "  'agreeableness: 0.46',\n",
       "  'agreeableness: 1.57',\n",
       "  'agreeableness: 0.3',\n",
       "  'agreeableness: -3.67',\n",
       "  'agreeableness: -0.97',\n",
       "  'agreeableness: -2.87',\n",
       "  'agreeableness: -2.08',\n",
       "  'agreeableness: -0.17',\n",
       "  'agreeableness: -1.29',\n",
       "  'agreeableness: -1.6',\n",
       "  'agreeableness: 0.94',\n",
       "  'agreeableness: 1.25',\n",
       "  'agreeableness: -1.92',\n",
       "  'agreeableness: 1.1',\n",
       "  'agreeableness: 1.89',\n",
       "  'agreeableness: 0.62',\n",
       "  'agreeableness: -1.44',\n",
       "  'agreeableness: -2.24',\n",
       "  'agreeableness: 1.73',\n",
       "  'agreeableness: -3.19'],\n",
       " 16: ['conscientiousness: -0.51',\n",
       "  'conscientiousness: 0.66',\n",
       "  'conscientiousness: 0.01',\n",
       "  'conscientiousness: -1.68',\n",
       "  'conscientiousness: 1.31',\n",
       "  'conscientiousness: -0.38',\n",
       "  'conscientiousness: 0.53',\n",
       "  'conscientiousness: -1.42',\n",
       "  'conscientiousness: -0.77',\n",
       "  'conscientiousness: 1.05',\n",
       "  'conscientiousness: 0.79',\n",
       "  'conscientiousness: -0.12',\n",
       "  'conscientiousness: -1.16',\n",
       "  'conscientiousness: 0.92',\n",
       "  'conscientiousness: -2.84',\n",
       "  'conscientiousness: 0.4',\n",
       "  'conscientiousness: -1.94',\n",
       "  'conscientiousness: -0.64',\n",
       "  'conscientiousness: 1.18',\n",
       "  'conscientiousness: 1.96',\n",
       "  'conscientiousness: 1.7',\n",
       "  'conscientiousness: -0.9',\n",
       "  'conscientiousness: -0.25',\n",
       "  'conscientiousness: 1.83',\n",
       "  'conscientiousness: 1.57',\n",
       "  'conscientiousness: 1.44',\n",
       "  'conscientiousness: -2.71',\n",
       "  'conscientiousness: -1.29',\n",
       "  'conscientiousness: 0.27',\n",
       "  'conscientiousness: 0.14']}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Anxiety disorder\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Anxiety disorder\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Leukocyte gene expression variation as a function of Big 5 dimensions of human personality\"\\n!Series_summary\\t\"Individual differences in basal leukocyte gene expression profiles as a function of Big 5 personality dimensions\"\\n!Series_overall_design\\t\"Gene expression profiling was carried out on peripheral blood mononuclear cell RNA samples collected from 119 healthy adults measured for the 5 major dimensions of human personality (Neuroticism, Extraversion, Openness, Agreeableness, Conscientiousness) using the NEO-FFI 60-item personality inventory.  Personality measures are z-score standardized.  Analyses control for major demographic characteristics (age, sex, Caucasian vs Non-Caucasian race) as well as Body Mass Index (BMI), smoking (CigDay), alcohol consumption (AlcDay), and physical activity (ExerDay, hours per day).  Additional secondary analyses controlled for the presence of minor physical symptoms (MinorSymptom, e.g.,hayfever, headache), medication use (BirthControl, Antidepressant, OtherMedication), and negative affective states (NegativeAffect, standardized values of PANAS Negative Affect scale).\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Experiment type: Risk prediction\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'age: 21\\', \\'age: 22\\', \\'age: 23\\', \\'age: 33\\', \\'age: 20\\', \\'age: 34\\', \\'age: 19\\', \\'age: 27\\', \\'age: 53\\', \\'age: 25\\', \\'age: 26\\', \\'age: 45\\', \\'age: 38\\', \\'age: 29\\', \\'age: 30\\', \\'age: 28\\', \\'age: 18\\', \\'age: 24\\', \\'age: 59\\', \\'age: 35\\', \\'age: 51\\', \\'age: 50\\', \\'age: 32\\'], 1: [\\'male: 0\\', \\'male: 1\\'], 2: [\\'bmi: 23\\', \\'bmi: 30\\', \\'bmi: 18\\', \\'bmi: 21.6\\', \\'bmi: 22\\', \\'bmi: 16\\', \\'bmi: 26\\', \\'bmi: 24\\', \\'bmi: 31\\', \\'bmi: 21\\', \\'bmi: missing\\', \\'bmi: 19\\', \\'bmi: 25\\', \\'bmi: 17.5\\', \\'bmi: 20\\', \\'bmi: 28\\', \\'bmi: 29\\', \\'bmi: 35\\', \\'bmi: 17\\', \\'bmi: 27\\', \\'bmi: 21.9\\', \\'bmi: 36\\'], 3: [\\'caucasian: 1\\', \\'caucasian: 0\\', \\'caucasian: missing\\'], 4: [\\'cigday: 0\\', \\'cigday: 7\\', \\'cigday: 12\\', \\'cigday: 3\\', \\'cigday: 5\\', \\'cigday: 0.1\\', \\'cigday: 1.4\\', \\'cigday: 20\\'], 5: [\\'alcday: 0.4\\', \\'alcday: 0\\', \\'alcday: 1.3\\', \\'alcday: 0.9\\', \\'alcday: 2\\', \\'alcday: 1\\', \\'alcday: 0.1\\', \\'alcday: 1.4\\', \\'alcday: 0.6\\', \\'alcday: 2.9\\', \\'alcday: 0.3\\', \\'alcday: 1.7\\', \\'alcday: 3\\', \\'alcday: 1.1\\', \\'alcday: 0.7\\', \\'alcday: 2.3\\', \\'alcday: 1.5\\', \\'alcday: 2.2\\', \\'alcday: 6\\', \\'alcday: 0.2\\'], 6: [\\'exerday: 0.5\\', \\'exerday: 0.9\\', \\'exerday: 0.6\\', \\'exerday: 1\\', \\'exerday: 0.1\\', \\'exerday: 0.3\\', \\'exerday: 0\\', \\'exerday: 2\\', \\'exerday: 1.5\\', \\'exerday: 0.7\\', \\'exerday: 0.4\\', \\'exerday: 0.8\\', \\'exerday: 0.75\\', \\'exerday: 1.4\\', \\'exerday: 3\\', \\'exerday: 1.1\\'], 7: [\\'minorsymptom: 0\\', \\'minorsymptom: missing\\', \\'minorsymptom: 1\\'], 8: [\\'birthcontrol: 0\\', \\'birthcontrol: 1\\', \\'birthcontrol: missing\\'], 9: [\\'antidepressant: 0\\', \\'antidepressant: 1\\', \\'antidepressant: missing\\'], 10: [\\'othermedication: 0\\', \\'othermedication: missing\\', \\'othermedication: 1\\'], 11: [\\'negativeaffect: 0.162\\', \\'negativeaffect: 1.319\\', \\'negativeaffect: 1.65\\', \\'negativeaffect: -0.83\\', \\'negativeaffect: 2.311\\', \\'negativeaffect: -0.664\\', \\'negativeaffect: -0.003\\', \\'negativeaffect: 2.972\\', \\'negativeaffect: 4.129\\', \\'negativeaffect: -0.334\\', \\'negativeaffect: -0.169\\', \\'negativeaffect: 0.658\\', \\'negativeaffect: 1.484\\', \\'negativeaffect: 0.823\\', \\'negativeaffect: 1.815\\', \\'negativeaffect: 2.476\\', \\'negativeaffect: 0.493\\', \\'negativeaffect: 0.988\\', \\'negativeaffect: 2.145\\'], 12: [\\'neuroticism: 0.18\\', \\'neuroticism: -0.96\\', \\'neuroticism: 2.58\\', \\'neuroticism: -1.47\\', \\'neuroticism: 1.06\\', \\'neuroticism: -0.46\\', \\'neuroticism: 0.3\\', \\'neuroticism: 0.94\\', \\'neuroticism: 2.46\\', \\'neuroticism: 0.81\\', \\'neuroticism: 0.43\\', \\'neuroticism: -1.09\\', \\'neuroticism: -0.08\\', \\'neuroticism: -0.71\\', \\'neuroticism: -0.2\\', \\'neuroticism: -1.34\\', \\'neuroticism: -0.84\\', \\'neuroticism: 1.19\\', \\'neuroticism: 0.56\\', \\'neuroticism: 0.05\\', \\'neuroticism: -0.33\\', \\'neuroticism: 1.95\\', \\'neuroticism: 2.08\\', \\'neuroticism: -0.58\\', \\'neuroticism: 0.68\\', \\'neuroticism: -1.97\\', \\'neuroticism: 1.44\\', \\'neuroticism: -1.85\\', \\'neuroticism: -1.22\\', \\'neuroticism: 1.32\\'], 13: [\\'extraversion: 0.46\\', \\'extraversion: -0.43\\', \\'extraversion: -1.01\\', \\'extraversion: 0.6\\', \\'extraversion: -0.72\\', \\'extraversion: -0.57\\', \\'extraversion: -1.16\\', \\'extraversion: 0.16\\', \\'extraversion: -1.75\\', \\'extraversion: 0.01\\', \\'extraversion: 0.9\\', \\'extraversion: 1.04\\', \\'extraversion: -1.31\\', \\'extraversion: 2.22\\', \\'extraversion: -1.6\\', \\'extraversion: 1.19\\', \\'extraversion: -2.34\\', \\'extraversion: -2.19\\', \\'extraversion: 1.34\\', \\'extraversion: 0.31\\', \\'extraversion: -0.28\\', \\'extraversion: -1.46\\', \\'extraversion: -0.87\\', \\'extraversion: -0.13\\', \\'extraversion: 1.93\\', \\'extraversion: 1.63\\', \\'extraversion: 2.07\\', \\'extraversion: 0.75\\', \\'extraversion: 1.49\\'], 14: [\\'openness: 0.27\\', \\'openness: 0.73\\', \\'openness: 2.14\\', \\'openness: -0.2\\', \\'openness: -1.14\\', \\'openness: 0.11\\', \\'openness: 0.89\\', \\'openness: 1.05\\', \\'openness: -0.36\\', \\'openness: -0.98\\', \\'openness: 1.67\\', \\'openness: -1.92\\', \\'openness: -0.83\\', \\'openness: -2.39\\', \\'openness: 1.52\\', \\'openness: -0.67\\', \\'openness: -0.05\\', \\'openness: 0.42\\', \\'openness: -0.52\\', \\'openness: 1.2\\', \\'openness: -1.77\\', \\'openness: 0.58\\', \\'openness: 2.45\\', \\'openness: -1.3\\', \\'openness: 1.36\\', \\'openness: 1.83\\', \\'openness: -1.61\\', \\'openness: -2.23\\', \\'openness: -2.08\\', \\'openness: -2.7\\'], 15: [\\'agreeableness: -0.02\\', \\'agreeableness: -0.81\\', \\'agreeableness: -0.49\\', \\'agreeableness: -0.33\\', \\'agreeableness: 0.78\\', \\'agreeableness: -1.13\\', \\'agreeableness: 0.14\\', \\'agreeableness: 1.41\\', \\'agreeableness: 0.46\\', \\'agreeableness: 1.57\\', \\'agreeableness: 0.3\\', \\'agreeableness: -3.67\\', \\'agreeableness: -0.97\\', \\'agreeableness: -2.87\\', \\'agreeableness: -2.08\\', \\'agreeableness: -0.17\\', \\'agreeableness: -1.29\\', \\'agreeableness: -1.6\\', \\'agreeableness: 0.94\\', \\'agreeableness: 1.25\\', \\'agreeableness: -1.92\\', \\'agreeableness: 1.1\\', \\'agreeableness: 1.89\\', \\'agreeableness: 0.62\\', \\'agreeableness: -1.44\\', \\'agreeableness: -2.24\\', \\'agreeableness: 1.73\\', \\'agreeableness: -3.19\\'], 16: [\\'conscientiousness: -0.51\\', \\'conscientiousness: 0.66\\', \\'conscientiousness: 0.01\\', \\'conscientiousness: -1.68\\', \\'conscientiousness: 1.31\\', \\'conscientiousness: -0.38\\', \\'conscientiousness: 0.53\\', \\'conscientiousness: -1.42\\', \\'conscientiousness: -0.77\\', \\'conscientiousness: 1.05\\', \\'conscientiousness: 0.79\\', \\'conscientiousness: -0.12\\', \\'conscientiousness: -1.16\\', \\'conscientiousness: 0.92\\', \\'conscientiousness: -2.84\\', \\'conscientiousness: 0.4\\', \\'conscientiousness: -1.94\\', \\'conscientiousness: -0.64\\', \\'conscientiousness: 1.18\\', \\'conscientiousness: 1.96\\', \\'conscientiousness: 1.7\\', \\'conscientiousness: -0.9\\', \\'conscientiousness: -0.25\\', \\'conscientiousness: 1.83\\', \\'conscientiousness: 1.57\\', \\'conscientiousness: 1.44\\', \\'conscientiousness: -2.71\\', \\'conscientiousness: -1.29\\', \\'conscientiousness: 0.27\\', \\'conscientiousness: 0.14\\']}\\n'"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 9\n",
    "age_row = 0\n",
    "gender_row = 1\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'antidepressant: 0':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if gender_string.lower() == 'male: 0':\n",
    "        return 0\n",
    "    elif gender_string.lower() == 'male: 1':\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_20364\\2239879167.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_20364\\2239879167.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "C:\\Users\\legion\\AppData\\Local\\Temp\\ipykernel_20364\\2239879167.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1481100</th>\n",
       "      <th>GSM1481101</th>\n",
       "      <th>GSM1481102</th>\n",
       "      <th>GSM1481103</th>\n",
       "      <th>GSM1481104</th>\n",
       "      <th>GSM1481105</th>\n",
       "      <th>GSM1481106</th>\n",
       "      <th>GSM1481107</th>\n",
       "      <th>GSM1481108</th>\n",
       "      <th>GSM1481109</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1481209</th>\n",
       "      <th>GSM1481210</th>\n",
       "      <th>GSM1481211</th>\n",
       "      <th>GSM1481212</th>\n",
       "      <th>GSM1481213</th>\n",
       "      <th>GSM1481214</th>\n",
       "      <th>GSM1481215</th>\n",
       "      <th>GSM1481216</th>\n",
       "      <th>GSM1481217</th>\n",
       "      <th>GSM1481218</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>21</td>\n",
       "      <td>22</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>22</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>22</td>\n",
       "      <td>32</td>\n",
       "      <td>27</td>\n",
       "      <td>22</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 119 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM1481100  GSM1481101  GSM1481102  GSM1481103  GSM1481104  \\\n",
       "Anxiety disorder           1           1           1           1           1   \n",
       "Age                       21          22          21          21          22   \n",
       "Gender                     0           1           0           1           0   \n",
       "\n",
       "                  GSM1481105  GSM1481106  GSM1481107  GSM1481108  GSM1481109  \\\n",
       "Anxiety disorder           1           1           0           1           0   \n",
       "Age                       23          23          23          23          23   \n",
       "Gender                     0           0           0           0           0   \n",
       "\n",
       "                  ...  GSM1481209  GSM1481210  GSM1481211  GSM1481212  \\\n",
       "Anxiety disorder  ...           1           1           1           1   \n",
       "Age               ...          24          20          26          22   \n",
       "Gender            ...           1           1           0           0   \n",
       "\n",
       "                  GSM1481213  GSM1481214  GSM1481215  GSM1481216  GSM1481217  \\\n",
       "Anxiety disorder           1           0           1           1           1   \n",
       "Age                       32          27          22          26          18   \n",
       "Gender                     1           1           1           1           1   \n",
       "\n",
       "                  GSM1481218  \n",
       "Anxiety disorder           1  \n",
       "Age                       20  \n",
       "Gender                     0  \n",
       "\n",
       "[3 rows x 119 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1481100</th>\n",
       "      <th>GSM1481101</th>\n",
       "      <th>GSM1481102</th>\n",
       "      <th>GSM1481103</th>\n",
       "      <th>GSM1481104</th>\n",
       "      <th>GSM1481105</th>\n",
       "      <th>GSM1481106</th>\n",
       "      <th>GSM1481107</th>\n",
       "      <th>GSM1481108</th>\n",
       "      <th>GSM1481109</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1481209</th>\n",
       "      <th>GSM1481210</th>\n",
       "      <th>GSM1481211</th>\n",
       "      <th>GSM1481212</th>\n",
       "      <th>GSM1481213</th>\n",
       "      <th>GSM1481214</th>\n",
       "      <th>GSM1481215</th>\n",
       "      <th>GSM1481216</th>\n",
       "      <th>GSM1481217</th>\n",
       "      <th>GSM1481218</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7A5</th>\n",
       "      <td>122.4265</td>\n",
       "      <td>131.7769</td>\n",
       "      <td>119.4328</td>\n",
       "      <td>118.3859</td>\n",
       "      <td>119.1651</td>\n",
       "      <td>127.3848</td>\n",
       "      <td>127.8088</td>\n",
       "      <td>124.0155</td>\n",
       "      <td>122.6773</td>\n",
       "      <td>123.7244</td>\n",
       "      <td>...</td>\n",
       "      <td>126.0626</td>\n",
       "      <td>133.5475</td>\n",
       "      <td>129.8606</td>\n",
       "      <td>120.7393</td>\n",
       "      <td>126.4916</td>\n",
       "      <td>137.7973</td>\n",
       "      <td>129.5558</td>\n",
       "      <td>121.2862</td>\n",
       "      <td>139.3874</td>\n",
       "      <td>122.1245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>878.7355</td>\n",
       "      <td>886.0469</td>\n",
       "      <td>949.6923</td>\n",
       "      <td>747.9395</td>\n",
       "      <td>874.6235</td>\n",
       "      <td>653.9559</td>\n",
       "      <td>831.1398</td>\n",
       "      <td>722.0098</td>\n",
       "      <td>1030.2450</td>\n",
       "      <td>832.8301</td>\n",
       "      <td>...</td>\n",
       "      <td>1334.3190</td>\n",
       "      <td>752.4141</td>\n",
       "      <td>907.3934</td>\n",
       "      <td>941.5509</td>\n",
       "      <td>1106.6530</td>\n",
       "      <td>1092.7820</td>\n",
       "      <td>1272.8460</td>\n",
       "      <td>1060.9150</td>\n",
       "      <td>936.6741</td>\n",
       "      <td>1060.6010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>130.8007</td>\n",
       "      <td>132.5247</td>\n",
       "      <td>141.0343</td>\n",
       "      <td>138.1274</td>\n",
       "      <td>144.6797</td>\n",
       "      <td>130.9009</td>\n",
       "      <td>143.6457</td>\n",
       "      <td>144.2968</td>\n",
       "      <td>140.4412</td>\n",
       "      <td>135.7075</td>\n",
       "      <td>...</td>\n",
       "      <td>139.9214</td>\n",
       "      <td>136.9379</td>\n",
       "      <td>142.3792</td>\n",
       "      <td>151.1532</td>\n",
       "      <td>148.4035</td>\n",
       "      <td>154.3208</td>\n",
       "      <td>143.4808</td>\n",
       "      <td>139.8078</td>\n",
       "      <td>134.4173</td>\n",
       "      <td>134.2816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A26C3</th>\n",
       "      <td>119.3683</td>\n",
       "      <td>114.8981</td>\n",
       "      <td>118.6767</td>\n",
       "      <td>112.5082</td>\n",
       "      <td>108.6636</td>\n",
       "      <td>116.0207</td>\n",
       "      <td>113.3711</td>\n",
       "      <td>116.4169</td>\n",
       "      <td>113.9820</td>\n",
       "      <td>112.3190</td>\n",
       "      <td>...</td>\n",
       "      <td>118.6542</td>\n",
       "      <td>118.0074</td>\n",
       "      <td>123.5586</td>\n",
       "      <td>111.4767</td>\n",
       "      <td>120.3181</td>\n",
       "      <td>118.5838</td>\n",
       "      <td>120.2557</td>\n",
       "      <td>123.6906</td>\n",
       "      <td>115.9226</td>\n",
       "      <td>119.2732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2BP1</th>\n",
       "      <td>119.3083</td>\n",
       "      <td>117.0984</td>\n",
       "      <td>115.3259</td>\n",
       "      <td>120.6321</td>\n",
       "      <td>117.4373</td>\n",
       "      <td>108.4148</td>\n",
       "      <td>115.9964</td>\n",
       "      <td>113.0969</td>\n",
       "      <td>119.0173</td>\n",
       "      <td>117.6048</td>\n",
       "      <td>...</td>\n",
       "      <td>121.1456</td>\n",
       "      <td>118.6000</td>\n",
       "      <td>119.0057</td>\n",
       "      <td>127.2296</td>\n",
       "      <td>119.4768</td>\n",
       "      <td>117.4570</td>\n",
       "      <td>116.3420</td>\n",
       "      <td>119.2073</td>\n",
       "      <td>122.7845</td>\n",
       "      <td>121.0886</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 119 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GSM1481100  GSM1481101  GSM1481102  GSM1481103  GSM1481104  GSM1481105  \\\n",
       "ID                                                                              \n",
       "7A5      122.4265    131.7769    119.4328    118.3859    119.1651    127.3848   \n",
       "A1BG     878.7355    886.0469    949.6923    747.9395    874.6235    653.9559   \n",
       "A1CF     130.8007    132.5247    141.0343    138.1274    144.6797    130.9009   \n",
       "A26C3    119.3683    114.8981    118.6767    112.5082    108.6636    116.0207   \n",
       "A2BP1    119.3083    117.0984    115.3259    120.6321    117.4373    108.4148   \n",
       "\n",
       "       GSM1481106  GSM1481107  GSM1481108  GSM1481109  ...  GSM1481209  \\\n",
       "ID                                                     ...               \n",
       "7A5      127.8088    124.0155    122.6773    123.7244  ...    126.0626   \n",
       "A1BG     831.1398    722.0098   1030.2450    832.8301  ...   1334.3190   \n",
       "A1CF     143.6457    144.2968    140.4412    135.7075  ...    139.9214   \n",
       "A26C3    113.3711    116.4169    113.9820    112.3190  ...    118.6542   \n",
       "A2BP1    115.9964    113.0969    119.0173    117.6048  ...    121.1456   \n",
       "\n",
       "       GSM1481210  GSM1481211  GSM1481212  GSM1481213  GSM1481214  GSM1481215  \\\n",
       "ID                                                                              \n",
       "7A5      133.5475    129.8606    120.7393    126.4916    137.7973    129.5558   \n",
       "A1BG     752.4141    907.3934    941.5509   1106.6530   1092.7820   1272.8460   \n",
       "A1CF     136.9379    142.3792    151.1532    148.4035    154.3208    143.4808   \n",
       "A26C3    118.0074    123.5586    111.4767    120.3181    118.5838    120.2557   \n",
       "A2BP1    118.6000    119.0057    127.2296    119.4768    117.4570    116.3420   \n",
       "\n",
       "       GSM1481216  GSM1481217  GSM1481218  \n",
       "ID                                         \n",
       "7A5      121.2862    139.3874    122.1245  \n",
       "A1BG    1060.9150    936.6741   1060.6010  \n",
       "A1CF     139.8078    134.4173    134.2816  \n",
       "A26C3    123.6906    115.9226    119.2732  \n",
       "A2BP1    119.2073    122.7845    121.0886  \n",
       "\n",
       "[5 rows x 119 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['7A5',\n",
       " 'A1BG',\n",
       " 'A1CF',\n",
       " 'A26C3',\n",
       " 'A2BP1',\n",
       " 'A2LD1',\n",
       " 'A2M',\n",
       " 'A2ML1',\n",
       " 'A3GALT2',\n",
       " 'A4GALT',\n",
       " 'A4GNT',\n",
       " 'AAA1',\n",
       " 'AAAS',\n",
       " 'AACS',\n",
       " 'AACSL',\n",
       " 'AADAC',\n",
       " 'AADACL1',\n",
       " 'AADACL2',\n",
       " 'AADACL3',\n",
       " 'AADACL4']"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['7A5', 'A1BG', 'A1CF', 'A26C3', 'A2BP1', 'A2LD1', 'A2M', 'A2ML1', 'A3GALT2', 'A4GALT', 'A4GNT', 'AAA1', 'AAAS', 'AACS', 'AACSL', 'AADAC', 'AADACL1', 'AADACL2', 'AADACL3', 'AADACL4']\\n\""
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'UCSC_RefGene_Name'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13 input query terms found dup hits:\t[('ABCC13', 2), ('ABCC6P1', 2), ('ABCC6P2', 3), ('ADAM3A', 2), ('ADAM6', 3), ('AGAP11', 2), ('AK2P2'\n",
      "95 input query terms found no hit:\t['7A5', 'A26C3', 'A2BP1', 'A2LD1', 'AACSL', 'AADACL1', 'AARS', 'ABCA11', 'ABHD7', 'ABHD9', 'ABP1', '\n",
      "6 input query terms found dup hits:\t[('ATP6AP1L', 2), ('ATXN8OS', 2), ('BAGE2', 2), ('BIRC8', 2), ('BRD7P3', 2), ('BRI3P1', 2)]\n",
      "370 input query terms found no hit:\t['ARMC4', 'ARMET', 'ARNTL', 'ARNTL2', 'ARP11', 'ARPM1', 'ARPM2', 'ARPP-21', 'ARS2', 'ARSE', 'ARVP612\n",
      "812 input query terms found no hit:\t['C15ORF51', 'C15ORF52', 'C15ORF53', 'C15ORF54', 'C15ORF55', 'C15ORF56', 'C15ORF57', 'C15ORF58', 'C1\n",
      "7 input query terms found dup hits:\t[('CATSPER2P1', 2), ('CCDC144NL', 2), ('CCNYL2', 2), ('CCNYL3', 2), ('CCT6P1', 2), ('CDR1', 2), ('CE\n",
      "126 input query terms found no hit:\t['CA5BP', 'CABC1', 'CAMSAP1L1', 'CARD17', 'CARKD', 'CARS', 'CART1', 'CASC1', 'CASC4', 'CASC5', 'CASR\n",
      "4 input query terms found dup hits:\t[('CLEC4GP1', 2), ('COPG2IT1', 2), ('CXADRP2', 3), ('CYP4Z2P', 2)]\n",
      "119 input query terms found no hit:\t['CLECL1', 'CLK2P', 'CLLU1OS', 'CLRN1OS', 'CMAH', 'CMYA3', 'CNIH', 'CNO', 'CNTD2', 'COBRA1', 'COL28'\n",
      "14 input query terms found dup hits:\t[('DEFB117', 2), ('DEFB122', 2), ('DHRS4L1', 3), ('DIRC1', 2), ('DLEU2L', 2), ('DNAJB3', 2), ('DNM1P\n",
      "135 input query terms found no hit:\t['DEFB109', 'DEFB109P1', 'DEFB130', 'DEFB131', 'DEFB137', 'DEFB32', 'DEFB4', 'DEM1', 'DEPDC6', 'DFNA\n",
      "7 input query terms found dup hits:\t[('FAM197Y2', 3), ('FAM41AY2', 2), ('FAM74A3', 2), ('FAM74A4', 2), ('FAM90A2P', 2), ('FAM90A6P', 3),\n",
      "494 input query terms found no hit:\t['EXOC3L', 'EXOD1', 'FABP5L2', 'FABP5L3', 'FABP5L7', 'FAIM3', 'FALZ', 'FAM100A', 'FAM100B', 'FAM101A\n",
      "10 input query terms found dup hits:\t[('FMO6P', 2), ('FMO9P', 2), ('FOLH1B', 2), ('GABPAP', 2), ('GAGE12B', 2), ('GGNBP1', 2), ('GGT8P', \n",
      "223 input query terms found no hit:\t['FLJ41821', 'FLJ41841', 'FLJ41856', 'FLJ41941', 'FLJ42102', 'FLJ42133', 'FLJ42177', 'FLJ42220', 'FL\n",
      "12 input query terms found dup hits:\t[('GTF2H2B', 2), ('GTF2IP1', 2), ('GUCY1B2', 2), ('HAR1A', 2), ('HBBP1', 2), ('HBD', 2), ('HCG4', 9)\n",
      "455 input query terms found no hit:\t['GRAMD2', 'GRAMD3', 'GRASP', 'GRINL1A', 'GRLF1', 'GRRP1', 'GS85', 'GSDM1', 'GSG2', 'GSTM1L', 'GSTTP\n",
      "1000 input query terms found no hit:\t['HS.146013', 'HS.146184', 'HS.146561', 'HS.146586', 'HS.146595', 'HS.146614', 'HS.146724', 'HS.1467\n",
      "1000 input query terms found no hit:\t['HS.526948', 'HS.527071', 'HS.527174', 'HS.527194', 'HS.527241', 'HS.527387', 'HS.527515', 'HS.5275\n",
      "982 input query terms found no hit:\t['HS.561625', 'HS.561632', 'HS.561633', 'HS.561644', 'HS.561656', 'HS.561679', 'HS.561735', 'HS.5617\n",
      "7 input query terms found dup hits:\t[('HSD17B7P2', 2), ('HSP90AB2P', 2), ('HSP90AB4P', 2), ('HSP90AB6P', 2), ('ID2B', 2), ('IFITM4P', 7)\n",
      "239 input query terms found no hit:\t['HSFYP1', 'HSGT1', 'HSN2', 'HSPBL2', 'HSPC047', 'HSPC072', 'HSPC111', 'HSPC157', 'HSPC159', 'HSPC17\n",
      "5 input query terms found dup hits:\t[('KIR3DX1', 5), ('KLKP1', 2), ('KRT126P', 2), ('KRT18P19', 2), ('KRT18P40', 3)]\n",
      "370 input query terms found no hit:\t['KIAA2018', 'KIAA2022', 'KIAA2026', 'KILLIN', 'KIRREL', 'KLHDC5', 'KLKBL4', 'KLRA1', 'KLRAQ1', 'KRT\n",
      "947 input query terms found no hit:\t['LOC100129067', 'LOC100129070', 'LOC100129076', 'LOC100129077', 'LOC100129081', 'LOC100129083', 'LO\n",
      "976 input query terms found no hit:\t['LOC100132329', 'LOC100132330', 'LOC100132336', 'LOC100132339', 'LOC100132340', 'LOC100132345', 'LO\n",
      "931 input query terms found no hit:\t['LOC147710', 'LOC147727', 'LOC147804', 'LOC147975', 'LOC148003', 'LOC148137', 'LOC148203', 'LOC1484\n",
      "935 input query terms found no hit:\t['LOC400968', 'LOC400986', 'LOC400987', 'LOC401002', 'LOC401007', 'LOC401010', 'LOC401019', 'LOC4010\n",
      "966 input query terms found no hit:\t['LOC642693', 'LOC642698', 'LOC642701', 'LOC642702', 'LOC642707', 'LOC642711', 'LOC642712', 'LOC6427\n",
      "955 input query terms found no hit:\t['LOC644907', 'LOC644908', 'LOC644912', 'LOC644914', 'LOC644916', 'LOC644917', 'LOC644919', 'LOC6449\n",
      "994 input query terms found no hit:\t['LOC647407', 'LOC647413', 'LOC647415', 'LOC647417', 'LOC647435', 'LOC647436', 'LOC647439', 'LOC6474\n",
      "992 input query terms found no hit:\t['LOC650995', 'LOC651004', 'LOC651006', 'LOC651007', 'LOC651008', 'LOC651012', 'LOC651017', 'LOC6510\n",
      "955 input query terms found no hit:\t['LOC653683', 'LOC653684', 'LOC653686', 'LOC653687', 'LOC653689', 'LOC653692', 'LOC653695', 'LOC6536\n",
      "4 input query terms found dup hits:\t[('LPAL2', 2), ('LRP5L', 2), ('LY6G6E', 8), ('LYPLA2P1', 6)]\n",
      "498 input query terms found no hit:\t['LOC729789', 'LOC729792', 'LOC729793', 'LOC729798', 'LOC729799', 'LOC729806', 'LOC729809', 'LOC7298\n",
      "4 input query terms found dup hits:\t[('MEG8', 2), ('MEIS3P1', 2), ('MIR320B2', 2), ('MMD2', 2)]\n",
      "181 input query terms found no hit:\t['MCMDC1', 'MDS1', 'MESDC1', 'MESDC2', 'METT10D', 'METT11D1', 'METT5D1', 'METTL10', 'METTL11A', 'MET\n",
      "14 input query terms found dup hits:\t[('MORF4', 2), ('MRS2P2', 2), ('MSTO2P', 2), ('MSX2P1', 2), ('MT1IP', 2), ('MT1JP', 2), ('MT1L', 2),\n",
      "118 input query terms found no hit:\t['MORG1', 'MOSC1', 'MOSC2', 'MOXD2', 'MPP5', 'MPP6', 'MRC1L1', 'MRE11A', 'MRLC2', 'MRP63', 'MRPS36',\n",
      "13 input query terms found dup hits:\t[('NUDT9P1', 2), ('NXF4', 2), ('NXF5', 2), ('OFCC1', 2), ('OR10J3', 3), ('OR2A9P', 2), ('OR6W1P', 3)\n",
      "70 input query terms found no hit:\t['NRD1', 'NS3BP', 'NSBP1', 'NSUN5B', 'NSUN5C', 'NT5C3', 'NT5C3L', 'NTF5', 'NUDT16P', 'NUDT4P1', 'NUP\n",
      "14 input query terms found dup hits:\t[('PCDHGB8P', 2), ('PDZK1P1', 2), ('PGM5P2', 2), ('PI4KAP1', 2), ('PKD1P1', 2), ('PLGLA', 2), ('PMCH\n",
      "95 input query terms found no hit:\t['PCDP1', 'PCNX', 'PCNXL2', 'PCNXL3', 'PCTK1', 'PCTK2', 'PCTK3', 'PDDC1', 'PDIA3P', 'PDXDC2', 'PDZD3\n",
      "17 input query terms found dup hits:\t[('PRKY', 2), ('PSORS1C3', 8), ('PTENP1', 2), ('PVT1', 2), ('PYY2', 2), ('RAB1C', 2), ('RAET1K', 2),\n",
      "93 input query terms found no hit:\t['PRIM2A', 'PRKCABP', 'PRKCB1', 'PRKCDBP', 'PRKRIR', 'PRMT10', 'PRNPIP', 'PRO0132', 'PRO0611', 'PRO0\n",
      "22 input query terms found dup hits:\t[('RNF126P1', 2), ('RNF138P1', 2), ('RNF5P1', 2), ('RP9P', 2), ('RPL13AP3', 2), ('RPL23AP53', 2), ('\n",
      "129 input query terms found no hit:\t['RNASEN', 'RNF12', 'RNF144', 'RNF160', 'RNF165', 'RNF190', 'RNF216L', 'RNF219', 'RNMTL1', 'RNPC2', \n",
      "29 input query terms found dup hits:\t[('SIGLEC5', 2), ('SLC6A10P', 2), ('SNHG6', 2), ('SNHG8', 2), ('SNHG9', 3), ('SNORA16A', 2), ('SNORA\n",
      "48 input query terms found no hit:\t['SHRM', 'SIGLECP3', 'SIKE', 'SILV', 'SIP1', 'SIVA', 'SKCG-1', 'SKIP', 'SKIV2L', 'SKIV2L2', 'SKP1A',\n",
      "12 input query terms found dup hits:\t[('SPRR2C', 2), ('SRP14P1', 2), ('STAG3L1', 2), ('STAG3L2', 2), ('STAG3L3', 2), ('STAG3L4', 2), ('SU\n",
      "91 input query terms found no hit:\t['SPAG4L', 'SPANXB2', 'SPANXE', 'SPANXF1', 'SPATA5', 'SPATA5L1', 'SPERT', 'SPG20', 'SPG3A', 'SPHAR',\n",
      "11 input query terms found dup hits:\t[('TMEM191A', 2), ('TNXA', 5), ('TOP1P1', 2), ('TPRXL', 2), ('TRIM16L', 2), ('TRPC2', 2), ('TTTY11',\n",
      "114 input query terms found no hit:\t['TINP1', 'TM7SF4', 'TMCO7', 'TMED10P', 'TMEM1', 'TMEM103', 'TMEM110', 'TMEM111', 'TMEM118', 'TMEM13\n",
      "10 input query terms found dup hits:\t[('UBE2MP1', 2), ('UBTFL5', 2), ('UOX', 2), ('VENTXP1', 2), ('VENTXP7', 2), ('WASH2P', 2), ('WASH3P'\n",
      "129 input query terms found no hit:\t['U1SNRNPBP', 'U2AF1L2', 'U58', 'UBE1', 'UBE1C', 'UBE1DC1', 'UBE2CBP', 'UBE2QP2', 'UBXD7', 'UCHL5IP'\n",
      "3 input query terms found dup hits:\t[('ZNF658B', 2), ('ZNF702P', 2), ('ZNF781', 2)]\n",
      "50 input query terms found no hit:\t['ZNF137', 'ZNF167', 'ZNF187', 'ZNF192', 'ZNF193', 'ZNF204', 'ZNF238', 'ZNF252', 'ZNF259', 'ZNF271',\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 119 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Anxiety disorder', the least common label is '0.0' with 11 occurrences. This represents 9.24% of the dataset.\n",
      "The distribution of the feature 'Anxiety disorder' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 20.0\n",
      "  50% (Median): 22.0\n",
      "  75%: 24.0\n",
      "Min: 18.0\n",
      "Max: 59.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 35 occurrences. This represents 29.41% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE60491</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>119</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "1  GSE60491       True          True      False     True        True   \n",
       "\n",
       "   sample_size note  \n",
       "1          119       "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GSE60491'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSE60491</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>119</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "1  GSE60491       True          True      False     True        True   \n",
       "\n",
       "   sample_size note  \n",
       "1          119       "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.295070023Z",
     "start_time": "2024-01-10T21:40:21.338515691Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anxiety disorder</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>878.7355</td>\n",
       "      <td>130.8007</td>\n",
       "      <td>98.22960</td>\n",
       "      <td>120.8270</td>\n",
       "      <td>127.7400</td>\n",
       "      <td>107.5727</td>\n",
       "      <td>203.9451</td>\n",
       "      <td>...</td>\n",
       "      <td>143.9445</td>\n",
       "      <td>142.8054</td>\n",
       "      <td>137.5964</td>\n",
       "      <td>276.6226</td>\n",
       "      <td>286.5479</td>\n",
       "      <td>117.3781</td>\n",
       "      <td>243.3207</td>\n",
       "      <td>143.0981</td>\n",
       "      <td>126.1316</td>\n",
       "      <td>174.5524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>886.0469</td>\n",
       "      <td>132.5247</td>\n",
       "      <td>107.48750</td>\n",
       "      <td>112.3008</td>\n",
       "      <td>133.3795</td>\n",
       "      <td>102.2422</td>\n",
       "      <td>252.5950</td>\n",
       "      <td>...</td>\n",
       "      <td>141.7784</td>\n",
       "      <td>155.8628</td>\n",
       "      <td>132.0846</td>\n",
       "      <td>288.0483</td>\n",
       "      <td>256.2910</td>\n",
       "      <td>116.3579</td>\n",
       "      <td>252.8782</td>\n",
       "      <td>138.5269</td>\n",
       "      <td>117.8299</td>\n",
       "      <td>185.8810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>949.6923</td>\n",
       "      <td>141.0343</td>\n",
       "      <td>94.14699</td>\n",
       "      <td>118.4021</td>\n",
       "      <td>122.5672</td>\n",
       "      <td>105.2472</td>\n",
       "      <td>234.2607</td>\n",
       "      <td>...</td>\n",
       "      <td>143.6750</td>\n",
       "      <td>148.7864</td>\n",
       "      <td>128.6895</td>\n",
       "      <td>290.3251</td>\n",
       "      <td>250.7705</td>\n",
       "      <td>112.9105</td>\n",
       "      <td>236.6002</td>\n",
       "      <td>140.7121</td>\n",
       "      <td>112.9356</td>\n",
       "      <td>170.8842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>747.9395</td>\n",
       "      <td>138.1274</td>\n",
       "      <td>110.05560</td>\n",
       "      <td>118.7682</td>\n",
       "      <td>130.9674</td>\n",
       "      <td>112.5170</td>\n",
       "      <td>262.9443</td>\n",
       "      <td>...</td>\n",
       "      <td>147.4358</td>\n",
       "      <td>162.2560</td>\n",
       "      <td>131.9409</td>\n",
       "      <td>232.3967</td>\n",
       "      <td>275.3432</td>\n",
       "      <td>105.5575</td>\n",
       "      <td>237.3100</td>\n",
       "      <td>134.0878</td>\n",
       "      <td>113.5443</td>\n",
       "      <td>170.7890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>874.6235</td>\n",
       "      <td>144.6797</td>\n",
       "      <td>104.10300</td>\n",
       "      <td>119.3217</td>\n",
       "      <td>149.0589</td>\n",
       "      <td>109.5576</td>\n",
       "      <td>225.8657</td>\n",
       "      <td>...</td>\n",
       "      <td>128.2727</td>\n",
       "      <td>145.9251</td>\n",
       "      <td>136.5600</td>\n",
       "      <td>322.6942</td>\n",
       "      <td>289.0635</td>\n",
       "      <td>107.0224</td>\n",
       "      <td>241.3908</td>\n",
       "      <td>131.0249</td>\n",
       "      <td>103.9684</td>\n",
       "      <td>182.1901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 17884 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Anxiety disorder   Age  Gender      A1BG      A1CF        A2M     A2ML1  \\\n",
       "0               1.0  21.0     0.0  878.7355  130.8007   98.22960  120.8270   \n",
       "1               1.0  22.0     1.0  886.0469  132.5247  107.48750  112.3008   \n",
       "2               1.0  21.0     0.0  949.6923  141.0343   94.14699  118.4021   \n",
       "3               1.0  21.0     1.0  747.9395  138.1274  110.05560  118.7682   \n",
       "4               1.0  22.0     0.0  874.6235  144.6797  104.10300  119.3217   \n",
       "\n",
       "    A3GALT2    A4GALT     A4GNT  ...    ZWILCH     ZWINT      ZXDA      ZXDB  \\\n",
       "0  127.7400  107.5727  203.9451  ...  143.9445  142.8054  137.5964  276.6226   \n",
       "1  133.3795  102.2422  252.5950  ...  141.7784  155.8628  132.0846  288.0483   \n",
       "2  122.5672  105.2472  234.2607  ...  143.6750  148.7864  128.6895  290.3251   \n",
       "3  130.9674  112.5170  262.9443  ...  147.4358  162.2560  131.9409  232.3967   \n",
       "4  149.0589  109.5576  225.8657  ...  128.2727  145.9251  136.5600  322.6942   \n",
       "\n",
       "       ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "0  286.5479  117.3781  243.3207  143.0981  126.1316  174.5524  \n",
       "1  256.2910  116.3579  252.8782  138.5269  117.8299  185.8810  \n",
       "2  250.7705  112.9105  236.6002  140.7121  112.9356  170.8842  \n",
       "3  275.3432  105.5575  237.3100  134.0878  113.5443  170.7890  \n",
       "4  289.0635  107.0224  241.3908  131.0249  103.9684  182.1901  \n",
       "\n",
       "[5 rows x 17884 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.358779256Z",
     "start_time": "2024-01-10T21:40:27.271337819Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n",
    "\n",
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_batch_effect(X):\n",
    "    \"\"\"\n",
    "    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n",
    "\n",
    "    Args:\n",
    "    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n",
    "\n",
    "    Returns:\n",
    "    bool: True if a potential batch effect is detected, False otherwise.\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "\n",
    "    # Computing XX^T\n",
    "    XXt = np.dot(X, X.T)\n",
    "\n",
    "    # Compute the eigenvalues of XX^T\n",
    "    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n",
    "    eigen_values = sorted(eigen_values, reverse=True)\n",
    "\n",
    "    # Check for large gaps in the eigenvalues\n",
    "    for i in range(len(eigen_values) - 1):\n",
    "        gap = eigen_values[i] - eigen_values[i + 1]\n",
    "        if gap > 1 / n_samples:  # You may need to adjust this threshold\n",
    "            return True\n",
    "\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546590575Z",
     "start_time": "2024-01-10T21:40:27.298495045Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sparse_lmm import VariableSelection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546733889Z",
     "start_time": "2024-01-10T21:40:27.535249532Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Select appropriate models based on whether the dataset has batch effect.\n",
    "# We experiment on two models for each branch. We will decide which one to choose later.\n",
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n",
    "    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n",
    "    indices = np.arange(X.shape[0])\n",
    "    np.random.shuffle(indices)\n",
    "\n",
    "    fold_size = len(X) // k\n",
    "    performances = []\n",
    "\n",
    "    for i in range(k):\n",
    "        # Split data into train and test based on the current fold\n",
    "        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n",
    "        train_indices = np.setdiff1d(indices, test_indices)\n",
    "\n",
    "        X_train, X_test = X[train_indices], X[test_indices]\n",
    "        Y_train, Y_test = Y[train_indices], Y[test_indices]\n",
    "        Z_train, Z_test = Z[train_indices], Z[test_indices]\n",
    "\n",
    "        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n",
    "        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n",
    "\n",
    "        # model = model_constructor(**model_params)\n",
    "        model = ResidualizationRegressor(model_constructor, model_params)\n",
    "        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n",
    "        predictions = model.predict(normalized_X_test, normalized_Z_test)\n",
    "\n",
    "        if target_type == 'binary':\n",
    "            predictions = (predictions > 0.5).astype(int)\n",
    "            Y_test = (Y_test > 0.5).astype(int)\n",
    "            performance = accuracy_score(Y_test, predictions)\n",
    "        elif target_type == 'continuous':\n",
    "            performance = mean_squared_error(Y_test, predictions)\n",
    "\n",
    "        performances.append(performance)\n",
    "\n",
    "    cv_mean = np.mean(performances)\n",
    "    cv_std = np.std(performances)\n",
    "\n",
    "    if target_type == 'binary':\n",
    "        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% Â± {(cv_std * 100):.2f}%')\n",
    "    else:\n",
    "        print(f'The cross-validation MSE is {(cv_mean * 100):.2f} Â± {(cv_std * 100):.2f}')\n",
    "\n",
    "    return cv_mean, cv_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_data(X_train, X_test=None):\n",
    "    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n",
    "    and optionally the test data\"\"\"\n",
    "    mean = np.mean(X_train, axis=0)\n",
    "    std = np.std(X_train, axis=0)\n",
    "\n",
    "    # Handling columns with std = 0\n",
    "    std_no_zero = np.where(std == 0, 1, std)\n",
    "\n",
    "    # Normalize X_train\n",
    "    X_train_normalized = (X_train - mean) / std_no_zero\n",
    "    # Set normalized values to 0 where std was 0\n",
    "    X_train_normalized[:, std == 0] = 0\n",
    "\n",
    "    if X_test is not None:\n",
    "        X_test_normalized = (X_test - mean) / std_no_zero\n",
    "        X_test_normalized[:, std == 0] = 0\n",
    "    else:\n",
    "        X_test_normalized = None\n",
    "\n",
    "    return X_train_normalized, X_test_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ResidualizationRegressor:\n",
    "    def __init__(self, regression_model_constructor, params=None):\n",
    "        if params is None:\n",
    "            params = {}\n",
    "        self.regression_model = regression_model_constructor(**params)\n",
    "        self.beta_Z = None  # Coefficients for regression of Y on Z\n",
    "        self.beta_X = None  # Coefficients for regression of residual on X\n",
    "        self.neg_log_p_values = None  # Negative logarithm of p-values\n",
    "        self.p_values = None  # Actual p-values\n",
    "\n",
    "    def _reshape_data(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the data to ensure it's in the correct format (2D array).\n",
    "\n",
    "        :param data: The input data (can be 1D or 2D array).\n",
    "        :return: Reshaped 2D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 1:\n",
    "            return data.reshape(-1, 1)\n",
    "        return data\n",
    "\n",
    "    def _reshape_output(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the output data to ensure it's in the correct format (1D array).\n",
    "\n",
    "        :param data: The output data (can be 1D or 2D array).\n",
    "        :return: Reshaped 1D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 2 and data.shape[1] == 1:\n",
    "            return data.ravel()\n",
    "        return data\n",
    "    \n",
    "    def fit(self, X, Y, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Y = self._reshape_data(Y)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        # Step 1: Linear regression of Y on Z\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n",
    "        Y_hat = Z_ones @ self.beta_Z\n",
    "        e_Y = Y - Y_hat  # Residual of Y\n",
    "\n",
    "        # Step 2: Regress the residual on X using the included regression model\n",
    "        self.regression_model.fit(X, e_Y)\n",
    "\n",
    "        # Obtain coefficients from the regression model\n",
    "        if hasattr(self.regression_model, 'coef_'):\n",
    "            self.beta_X = self.regression_model.coef_\n",
    "        elif hasattr(self.regression_model, 'getBeta'):\n",
    "            beta_output = self.regression_model.getBeta()\n",
    "            self.beta_X = self._reshape_output(beta_output)\n",
    "\n",
    "        # Obtain negative logarithm of p-values, if available\n",
    "        if hasattr(self.regression_model, 'getNegLogP'):\n",
    "            neg_log_p_output = self.regression_model.getNegLogP()\n",
    "            if neg_log_p_output is not None:\n",
    "                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n",
    "                self.p_values = np.exp(-self.neg_log_p_values)\n",
    "                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n",
    "                p_values_Z = np.full(Z.shape[1], np.nan)\n",
    "                self.p_values = np.concatenate((p_values_Z, self.p_values))\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, mean_squared_error\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:30.570471636Z",
     "start_time": "2024-01-10T21:40:27.546295159Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "Module 'scipy' has no attribute 'exp'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\scipy\\__init__.py:150\u001b[0m, in \u001b[0;36m__getattr__\u001b[1;34m(name)\u001b[0m\n\u001b[0;32m    149\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 150\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mglobals\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[43mname\u001b[49m\u001b[43m]\u001b[49m\n\u001b[0;32m    151\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n",
      "\u001b[1;31mKeyError\u001b[0m: 'exp'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[98], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m trait_type \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbinary\u001b[39m\u001b[38;5;124m'\u001b[39m  \u001b[38;5;66;03m# Remember to set this properly, either 'binary' or 'continuous'\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m cv_mean1, cv_std1 \u001b[38;5;241m=\u001b[39m \u001b[43mcross_validation\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mY\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mZ\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel_constructor1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel_params1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtarget_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrait_type\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[94], line 23\u001b[0m, in \u001b[0;36mcross_validation\u001b[1;34m(X, Y, Z, model_constructor, model_params, k, target_type)\u001b[0m\n\u001b[0;32m     21\u001b[0m \u001b[38;5;66;03m# model = model_constructor(**model_params)\u001b[39;00m\n\u001b[0;32m     22\u001b[0m model \u001b[38;5;241m=\u001b[39m ResidualizationRegressor(model_constructor, model_params)\n\u001b[1;32m---> 23\u001b[0m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnormalized_X_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mY_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnormalized_Z_train\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     24\u001b[0m predictions \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39mpredict(normalized_X_test, normalized_Z_test)\n\u001b[0;32m     26\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m target_type \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbinary\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "Cell \u001b[1;32mIn[96], line 45\u001b[0m, in \u001b[0;36mResidualizationRegressor.fit\u001b[1;34m(self, X, Y, Z)\u001b[0m\n\u001b[0;32m     42\u001b[0m e_Y \u001b[38;5;241m=\u001b[39m Y \u001b[38;5;241m-\u001b[39m Y_hat  \u001b[38;5;66;03m# Residual of Y\u001b[39;00m\n\u001b[0;32m     44\u001b[0m \u001b[38;5;66;03m# Step 2: Regress the residual on X using the included regression model\u001b[39;00m\n\u001b[1;32m---> 45\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mregression_model\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43me_Y\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     47\u001b[0m \u001b[38;5;66;03m# Obtain coefficients from the regression model\u001b[39;00m\n\u001b[0;32m     48\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mregression_model, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcoef_\u001b[39m\u001b[38;5;124m'\u001b[39m):\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\sparse_lmm\\VariableSelection.py:45\u001b[0m, in \u001b[0;36mVariableSelection.fit\u001b[1;34m(self, X, y)\u001b[0m\n\u001b[0;32m     43\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfit\u001b[39m(\u001b[38;5;28mself\u001b[39m, X, y):\n\u001b[0;32m     44\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39macceleration:\n\u001b[1;32m---> 45\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit_fast\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     46\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m     47\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfit_detail(X, y)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\sparse_lmm\\VariableSelection.py:58\u001b[0m, in \u001b[0;36mVariableSelection.fit_fast\u001b[1;34m(self, X, y)\u001b[0m\n\u001b[0;32m     55\u001b[0m     y \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mreshape(y, (n_s, \u001b[38;5;241m1\u001b[39m))\n\u001b[0;32m     57\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39my \u001b[38;5;241m=\u001b[39m y\n\u001b[1;32m---> 58\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mS, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mU, ldelta0\u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain_nullmodel\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mK\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     60\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mldelta0 \u001b[38;5;241m=\u001b[39m ldelta0\n\u001b[0;32m     62\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodified:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\sparse_lmm\\VariableSelection.py:210\u001b[0m, in \u001b[0;36mVariableSelection.train_nullmodel\u001b[1;34m(self, y, K, S, U, numintervals)\u001b[0m\n\u001b[0;32m    208\u001b[0m ldeltagrid \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marange(numintervals \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m) \u001b[38;5;241m/\u001b[39m (numintervals \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m1.0\u001b[39m) \u001b[38;5;241m*\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mldeltamax \u001b[38;5;241m-\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mldeltamin) \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mldeltamin\n\u001b[0;32m    209\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m np\u001b[38;5;241m.\u001b[39marange(numintervals \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m):\n\u001b[1;32m--> 210\u001b[0m     nllgrid[i] \u001b[38;5;241m=\u001b[39m \u001b[43mnLLeval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mldeltagrid\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mUy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mS\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;66;03m# the method is in helpingMethods\u001b[39;00m\n\u001b[0;32m    212\u001b[0m nllmin \u001b[38;5;241m=\u001b[39m nllgrid\u001b[38;5;241m.\u001b[39mmin()\n\u001b[0;32m    213\u001b[0m ldeltaopt_glob \u001b[38;5;241m=\u001b[39m ldeltagrid[nllgrid\u001b[38;5;241m.\u001b[39margmin()]\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\sparse_lmm\\helpingMethods.py:72\u001b[0m, in \u001b[0;36mnLLeval\u001b[1;34m(ldelta, Uy, S, REML)\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     64\u001b[0m \u001b[38;5;124;03mevaluate the negative log likelihood of a random effects model:\u001b[39;00m\n\u001b[0;32m     65\u001b[0m \u001b[38;5;124;03mnLL = 1/2(n_s*log(2pi) + logdet(K) + 1/ss * y^T(K + deltaI)^{-1}y,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     69\u001b[0m \u001b[38;5;124;03mldelta: log-transformed ratio sigma_gg/sigma_ee\u001b[39;00m\n\u001b[0;32m     70\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     71\u001b[0m n_s \u001b[38;5;241m=\u001b[39m Uy\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m---> 72\u001b[0m delta \u001b[38;5;241m=\u001b[39m \u001b[43mscipy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexp\u001b[49m(ldelta)\n\u001b[0;32m     74\u001b[0m \u001b[38;5;66;03m# evaluate log determinant\u001b[39;00m\n\u001b[0;32m     75\u001b[0m Sd \u001b[38;5;241m=\u001b[39m S \u001b[38;5;241m+\u001b[39m delta\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\scipy\\__init__.py:152\u001b[0m, in \u001b[0;36m__getattr__\u001b[1;34m(name)\u001b[0m\n\u001b[0;32m    150\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mglobals\u001b[39m()[name]\n\u001b[0;32m    151\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n\u001b[1;32m--> 152\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\n\u001b[0;32m    153\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mModule \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mscipy\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m has no attribute \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    154\u001b[0m     )\n",
      "\u001b[1;31mAttributeError\u001b[0m: Module 'scipy' has no attribute 'exp'"
     ]
    }
   ],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:47.575878022Z",
     "start_time": "2024-01-10T21:40:30.570907992Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The cross-validation accuracy is 48.70% Â± 12.42%\n"
     ]
    }
   ],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.663737008Z",
     "start_time": "2024-01-10T21:40:47.576021186Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha for Lasso: 0.0003\n"
     ]
    }
   ],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n",
    "                     threshold: float = 0.05, save_output: bool = True,\n",
    "                     output_dir: str = './output', model_id: int = 1) -> None:\n",
    "    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n",
    "    consists of one variable about condition and multiple variables about genetic factors.\n",
    "    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n",
    "    p-values or non-zero coefficients, depending on the availability of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    model (Any): The trained regression Model.\n",
    "    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n",
    "    trait (str): The target trait of interest.\n",
    "    condition (str): The specific condition to examine within the model.\n",
    "    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n",
    "    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n",
    "    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n",
    "    model_id (int): The index of the model, 1 or 2.\n",
    "\n",
    "    Returns:\n",
    "    None: This function does not return anything but prints and optionally saves the output.\n",
    "    \"\"\"\n",
    "    coefficients = model.get_coefficients().reshape(-1).tolist()\n",
    "    p_values = model.get_p_values()\n",
    "    if p_values is None:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients\n",
    "        })\n",
    "    else:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients,\n",
    "            'p_value': p_values.reshape(-1).tolist()\n",
    "        })\n",
    "    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n",
    "\n",
    "    print(f\"Effect of the condition on the target variable:\")\n",
    "    print(f\"Variable: {condition}\")\n",
    "    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n",
    "    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n",
    "    if p_values is None:\n",
    "        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n",
    "        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n",
    "        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n",
    "            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n",
    "    else:\n",
    "        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n",
    "        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n",
    "        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n",
    "        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n",
    "        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n",
    "            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n",
    "\n",
    "    print(significant_genes_sorted.to_string(index=False))\n",
    "\n",
    "    # Optionally, save this to a CSV file\n",
    "    if save_output:\n",
    "        significant_genes_sorted.to_csv(\n",
    "            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.673565176Z",
     "start_time": "2024-01-10T21:40:51.665291317Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect of the condition on the target variable:\n",
      "Variable: Age\n",
      "Coefficient: -0.0515\n",
      "Found 53 genes with non-zero coefficients associated with the trait 'Anxiety disorder' conditional on the factor 'Age'. These genes are identified as significant based on the regression model.\n",
      "    Variable  Coefficient  Absolute Coefficient\n",
      "        BAG4     3.651880              3.651880\n",
      "       TMOD4    -3.645051              3.645051\n",
      "       TEKT5    -3.502467              3.502467\n",
      "       ADH1B    -3.340335              3.340335\n",
      "       BCAT1    -3.058921              3.058921\n",
      "      TTTY13    -2.910795              2.910795\n",
      "   KRTAP20-1     2.371920              2.371920\n",
      "       LRRN4     2.309673              2.309673\n",
      "       FBXL6    -2.205952              2.205952\n",
      "      GOLGB1    -1.840282              1.840282\n",
      "     ZDHHC24     1.769598              1.769598\n",
      "      TSPAN7    -1.746164              1.746164\n",
      "        RAG2    -1.686070              1.686070\n",
      "     SLC46A1     1.652568              1.652568\n",
      "LOC100130698    -1.569555              1.569555\n",
      "     TMEM125    -1.534171              1.534171\n",
      "      DMRTB1    -1.532772              1.532772\n",
      "     PLEKHH3     1.442971              1.442971\n",
      "       ABCB8     1.399242              1.399242\n",
      "     ZC3H11B    -1.345334              1.345334\n",
      "      CELA2B    -1.270608              1.270608\n",
      "      OR2AK2    -1.039929              1.039929\n",
      "         TNR     0.999594              0.999594\n",
      "       RBM19     0.921270              0.921270\n",
      "       OR2J3     0.777602              0.777602\n",
      "       P2RX2    -0.710358              0.710358\n",
      "        HCG4     0.710056              0.710056\n",
      "        MECR    -0.709885              0.709885\n",
      "       OR9G9     0.681016              0.681016\n",
      "        CSN2     0.665679              0.665679\n",
      "     MIR374A    -0.610289              0.610289\n",
      "       PTGES    -0.592740              0.592740\n",
      "       CORO6     0.588652              0.588652\n",
      "       PDRG1    -0.588185              0.588185\n",
      "       FRMD5    -0.573362              0.573362\n",
      "   LOC644554     0.480945              0.480945\n",
      "         ZP1     0.479285              0.479285\n",
      "      OR2AG2    -0.452852              0.452852\n",
      "       RAB26     0.445023              0.445023\n",
      "         DPT    -0.398954              0.398954\n",
      "      B3GNT7     0.394100              0.394100\n",
      "     PLA2G2E    -0.318673              0.318673\n",
      "        CRB3     0.248713              0.248713\n",
      "        IRS1     0.242018              0.242018\n",
      "        UBA6    -0.231945              0.231945\n",
      "      IGFBP3     0.216234              0.216234\n",
      "       MATN2     0.131439              0.131439\n",
      "       PPM1H    -0.102653              0.102653\n",
      "        PEF1    -0.080650              0.080650\n",
      "    C12orf54    -0.048741              0.048741\n",
      "       ORAI3    -0.016968              0.016968\n",
      "        HCN3    -0.001963              0.001963\n",
      "       WDR77    -0.001263              0.001263\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/hx/qdnjfcxj2kbfy5k46ywrbrz40000gn/T/ipykernel_99910/2152198575.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n"
     ]
    }
   ],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.730825913Z",
     "start_time": "2024-01-10T21:40:51.675049728Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect of the condition on the target variable:\n",
      "Variable: Age\n",
      "Coefficient: -0.0515\n",
      "Found 0 significant genes associated with the trait 'Anxiety disorder' conditional on the factor 'Age', with corrected p-value < 0.05:\n",
      "Empty DataFrame\n",
      "Columns: [Variable, Coefficient, p_value, corrected_p_value]\n",
      "Index: []\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/hx/qdnjfcxj2kbfy5k46ywrbrz40000gn/T/ipykernel_99910/2152198575.py:51: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"
     ]
    }
   ],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
