{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "# Set your preferred name\n",
    "USER = \"Jiayi\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output'\n",
    "TRAIT = 'Hepatitis'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.928143012Z",
     "start_time": "2024-01-10T21:37:25.918685567Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.554597756Z",
     "start_time": "2024-01-10T21:37:26.530982646Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.999521704Z",
     "start_time": "2024-01-10T21:37:26.987471072Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Adrenocortical_Cancer_(ACC)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_ACC_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.180581225Z",
     "start_time": "2024-01-10T21:37:27.217363685Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.185775284Z",
     "start_time": "2024-01-10T21:37:30.181745846Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.269854441Z",
     "start_time": "2024-01-10T21:37:30.184225525Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270036796Z",
     "start_time": "2024-01-10T21:37:30.227180075Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = [ 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = [ 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270232386Z",
     "start_time": "2024-01-10T21:37:30.227282564Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270339062Z",
     "start_time": "2024-01-10T21:37:30.227412883Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270520195Z",
     "start_time": "2024-01-10T21:37:30.227467704Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270626721Z",
     "start_time": "2024-01-10T21:37:30.227530279Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270719762Z",
     "start_time": "2024-01-10T21:37:30.227585330Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.001977877Z",
     "start_time": "2024-01-10T21:37:30.227635342Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import mygene\n",
    "\n",
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.108889749Z",
     "start_time": "2024-01-10T21:38:18.021199718Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.112793384Z",
     "start_time": "2024-01-10T21:38:18.107415526Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.377898533Z",
     "start_time": "2024-01-10T21:38:18.111417452Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425336629Z",
     "start_time": "2024-01-10T21:38:31.419043732Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE113392', 'GSE113409', 'GSE140249', 'GSE78160', 'GSE85550']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Hepatitis\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:02.935079426Z",
     "start_time": "2024-01-10T21:39:02.923698385Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Distinct serum microRNA profiles in patients with active autoimmune hepatitis\"\n",
      "!Series_summary\t\"Serum microRNAs are useful for the early diagnosis and management of autoimmune hepatitis\"\n",
      "!Series_overall_design\t\"Serum microRNA profiles of patients with autoimmune hepatitis (AIH),  primary biliary cholangitis (PBC), overlap syndrome (OS), or without hepatic diseases.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM4158202</th>\n",
       "      <th>GSM4158203</th>\n",
       "      <th>GSM4158204</th>\n",
       "      <th>GSM4158205</th>\n",
       "      <th>GSM4158206</th>\n",
       "      <th>GSM4158207</th>\n",
       "      <th>GSM4158208</th>\n",
       "      <th>GSM4158209</th>\n",
       "      <th>GSM4158210</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4158334</th>\n",
       "      <th>GSM4158335</th>\n",
       "      <th>GSM4158336</th>\n",
       "      <th>GSM4158337</th>\n",
       "      <th>GSM4158338</th>\n",
       "      <th>GSM4158339</th>\n",
       "      <th>GSM4158340</th>\n",
       "      <th>GSM4158341</th>\n",
       "      <th>GSM4158342</th>\n",
       "      <th>GSM4158343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 32</td>\n",
       "      <td>age: 61</td>\n",
       "      <td>age: 74</td>\n",
       "      <td>age: 27</td>\n",
       "      <td>age: 72</td>\n",
       "      <td>age: 28</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 76</td>\n",
       "      <td>age: 38</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 55</td>\n",
       "      <td>age: 48</td>\n",
       "      <td>age: 35</td>\n",
       "      <td>age: 34</td>\n",
       "      <td>age: 34</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 52</td>\n",
       "      <td>age: 54</td>\n",
       "      <td>age: 59</td>\n",
       "      <td>age: 69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>disease state: AIH</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "      <td>disease state: Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>treatment: Pre</td>\n",
       "      <td>...</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "      <td>treatment: N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 143 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession          GSM4158202          GSM4158203  \\\n",
       "0  !Sample_characteristics_ch1              Sex: F              Sex: F   \n",
       "1  !Sample_characteristics_ch1             age: 32             age: 61   \n",
       "2  !Sample_characteristics_ch1  disease state: AIH  disease state: AIH   \n",
       "3  !Sample_characteristics_ch1      treatment: Pre      treatment: Pre   \n",
       "4  !Sample_characteristics_ch1       tissue: serum       tissue: serum   \n",
       "\n",
       "           GSM4158204          GSM4158205          GSM4158206  \\\n",
       "0              Sex: M              Sex: F              Sex: M   \n",
       "1             age: 74             age: 27             age: 72   \n",
       "2  disease state: AIH  disease state: AIH  disease state: AIH   \n",
       "3      treatment: Pre      treatment: Pre      treatment: Pre   \n",
       "4       tissue: serum       tissue: serum       tissue: serum   \n",
       "\n",
       "           GSM4158207          GSM4158208          GSM4158209  \\\n",
       "0              Sex: M              Sex: M              Sex: M   \n",
       "1             age: 28             age: 71             age: 76   \n",
       "2  disease state: AIH  disease state: AIH  disease state: AIH   \n",
       "3      treatment: Pre      treatment: Pre      treatment: Pre   \n",
       "4       tissue: serum       tissue: serum       tissue: serum   \n",
       "\n",
       "           GSM4158210  ...              GSM4158334              GSM4158335  \\\n",
       "0              Sex: F  ...                  Sex: M                  Sex: F   \n",
       "1             age: 38  ...                 age: 55                 age: 48   \n",
       "2  disease state: AIH  ...  disease state: Healthy  disease state: Healthy   \n",
       "3      treatment: Pre  ...          treatment: N/A          treatment: N/A   \n",
       "4       tissue: serum  ...           tissue: serum           tissue: serum   \n",
       "\n",
       "               GSM4158336              GSM4158337              GSM4158338  \\\n",
       "0                  Sex: F                  Sex: F                  Sex: F   \n",
       "1                 age: 35                 age: 34                 age: 34   \n",
       "2  disease state: Healthy  disease state: Healthy  disease state: Healthy   \n",
       "3          treatment: N/A          treatment: N/A          treatment: N/A   \n",
       "4           tissue: serum           tissue: serum           tissue: serum   \n",
       "\n",
       "               GSM4158339              GSM4158340              GSM4158341  \\\n",
       "0                  Sex: F                  Sex: F                  Sex: M   \n",
       "1                 age: 23                 age: 52                 age: 54   \n",
       "2  disease state: Healthy  disease state: Healthy  disease state: Healthy   \n",
       "3          treatment: N/A          treatment: N/A          treatment: N/A   \n",
       "4           tissue: serum           tissue: serum           tissue: serum   \n",
       "\n",
       "               GSM4158342              GSM4158343  \n",
       "0                  Sex: M                  Sex: F  \n",
       "1                 age: 59                 age: 69  \n",
       "2  disease state: Healthy  disease state: Healthy  \n",
       "3          treatment: N/A          treatment: N/A  \n",
       "4           tissue: serum           tissue: serum  \n",
       "\n",
       "[5 rows x 143 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No gene mapping\n",
    "cohort = accession_num = \"GSE140249\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease state: AIH',\n",
       "       'disease state: PBC', 'disease state: OS',\n",
       "       'disease state: Healthy'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = 1\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if (tissue_type == 'disease state: AIH' or tissue_type == 'disease state: OS'):\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# 这个函数的目的是将年龄的字符串表示转换为一个连续的数值型表示。如果年龄未知（例如，标记为'n.a.'），则返回None。\n",
    "# 函数尝试从传入的字符串中提取出一个整数作为年龄值。如果字符串的格式不符合预期，导致提取失败，同样返回None。\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# 这个函数将性别的字符串表示转换为二进制值，其中“female”对应1，“male”对应0。如果性别未知或字符串不符合预期格式，则返回None。\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM4158202</th>\n",
       "      <th>GSM4158203</th>\n",
       "      <th>GSM4158204</th>\n",
       "      <th>GSM4158205</th>\n",
       "      <th>GSM4158206</th>\n",
       "      <th>GSM4158207</th>\n",
       "      <th>GSM4158208</th>\n",
       "      <th>GSM4158209</th>\n",
       "      <th>GSM4158210</th>\n",
       "      <th>GSM4158211</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4158334</th>\n",
       "      <th>GSM4158335</th>\n",
       "      <th>GSM4158336</th>\n",
       "      <th>GSM4158337</th>\n",
       "      <th>GSM4158338</th>\n",
       "      <th>GSM4158339</th>\n",
       "      <th>GSM4158340</th>\n",
       "      <th>GSM4158341</th>\n",
       "      <th>GSM4158342</th>\n",
       "      <th>GSM4158343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hepatitis</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>32</td>\n",
       "      <td>61</td>\n",
       "      <td>74</td>\n",
       "      <td>27</td>\n",
       "      <td>72</td>\n",
       "      <td>28</td>\n",
       "      <td>71</td>\n",
       "      <td>76</td>\n",
       "      <td>38</td>\n",
       "      <td>41</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>48</td>\n",
       "      <td>35</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>23</td>\n",
       "      <td>52</td>\n",
       "      <td>54</td>\n",
       "      <td>59</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM4158202  GSM4158203  GSM4158204  GSM4158205  GSM4158206  \\\n",
       "Hepatitis           1           1           1           1           1   \n",
       "Age                32          61          74          27          72   \n",
       "Gender              1           1           0           1           0   \n",
       "\n",
       "           GSM4158207  GSM4158208  GSM4158209  GSM4158210  GSM4158211  ...  \\\n",
       "Hepatitis           1           1           1           1           1  ...   \n",
       "Age                28          71          76          38          41  ...   \n",
       "Gender              0           0           0           1           1  ...   \n",
       "\n",
       "           GSM4158334  GSM4158335  GSM4158336  GSM4158337  GSM4158338  \\\n",
       "Hepatitis           0           0           0           0           0   \n",
       "Age                55          48          35          34          34   \n",
       "Gender              0           1           1           1           1   \n",
       "\n",
       "           GSM4158339  GSM4158340  GSM4158341  GSM4158342  GSM4158343  \n",
       "Hepatitis           0           0           0           0           0  \n",
       "Age                23          52          54          59          69  \n",
       "Gender              1           1           0           0           1  \n",
       "\n",
       "[3 rows x 142 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM4158202</th>\n",
       "      <th>GSM4158203</th>\n",
       "      <th>GSM4158204</th>\n",
       "      <th>GSM4158205</th>\n",
       "      <th>GSM4158206</th>\n",
       "      <th>GSM4158207</th>\n",
       "      <th>GSM4158208</th>\n",
       "      <th>GSM4158209</th>\n",
       "      <th>GSM4158210</th>\n",
       "      <th>GSM4158211</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4158334</th>\n",
       "      <th>GSM4158335</th>\n",
       "      <th>GSM4158336</th>\n",
       "      <th>GSM4158337</th>\n",
       "      <th>GSM4158338</th>\n",
       "      <th>GSM4158339</th>\n",
       "      <th>GSM4158340</th>\n",
       "      <th>GSM4158341</th>\n",
       "      <th>GSM4158342</th>\n",
       "      <th>GSM4158343</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIMAT0000062</th>\n",
       "      <td>2.840</td>\n",
       "      <td>2.634</td>\n",
       "      <td>3.026</td>\n",
       "      <td>2.769</td>\n",
       "      <td>2.793</td>\n",
       "      <td>4.163</td>\n",
       "      <td>3.741</td>\n",
       "      <td>4.383</td>\n",
       "      <td>4.771</td>\n",
       "      <td>-4.833</td>\n",
       "      <td>...</td>\n",
       "      <td>3.863</td>\n",
       "      <td>3.344</td>\n",
       "      <td>4.191</td>\n",
       "      <td>4.145</td>\n",
       "      <td>4.225</td>\n",
       "      <td>3.173</td>\n",
       "      <td>3.030</td>\n",
       "      <td>4.526</td>\n",
       "      <td>2.904</td>\n",
       "      <td>6.561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0000063</th>\n",
       "      <td>3.267</td>\n",
       "      <td>3.881</td>\n",
       "      <td>3.554</td>\n",
       "      <td>3.367</td>\n",
       "      <td>3.403</td>\n",
       "      <td>4.514</td>\n",
       "      <td>4.381</td>\n",
       "      <td>4.681</td>\n",
       "      <td>4.846</td>\n",
       "      <td>-3.565</td>\n",
       "      <td>...</td>\n",
       "      <td>4.468</td>\n",
       "      <td>4.054</td>\n",
       "      <td>4.552</td>\n",
       "      <td>4.003</td>\n",
       "      <td>4.497</td>\n",
       "      <td>4.079</td>\n",
       "      <td>4.087</td>\n",
       "      <td>4.730</td>\n",
       "      <td>4.321</td>\n",
       "      <td>6.585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0000064</th>\n",
       "      <td>2.753</td>\n",
       "      <td>2.696</td>\n",
       "      <td>3.104</td>\n",
       "      <td>2.567</td>\n",
       "      <td>3.514</td>\n",
       "      <td>3.866</td>\n",
       "      <td>3.775</td>\n",
       "      <td>4.007</td>\n",
       "      <td>4.458</td>\n",
       "      <td>-4.833</td>\n",
       "      <td>...</td>\n",
       "      <td>3.674</td>\n",
       "      <td>3.512</td>\n",
       "      <td>4.011</td>\n",
       "      <td>3.327</td>\n",
       "      <td>4.058</td>\n",
       "      <td>3.185</td>\n",
       "      <td>3.097</td>\n",
       "      <td>4.573</td>\n",
       "      <td>3.593</td>\n",
       "      <td>6.395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0000065</th>\n",
       "      <td>2.615</td>\n",
       "      <td>2.536</td>\n",
       "      <td>2.677</td>\n",
       "      <td>2.441</td>\n",
       "      <td>2.997</td>\n",
       "      <td>3.737</td>\n",
       "      <td>3.148</td>\n",
       "      <td>4.080</td>\n",
       "      <td>4.290</td>\n",
       "      <td>-4.833</td>\n",
       "      <td>...</td>\n",
       "      <td>3.232</td>\n",
       "      <td>3.245</td>\n",
       "      <td>3.758</td>\n",
       "      <td>2.931</td>\n",
       "      <td>3.865</td>\n",
       "      <td>2.710</td>\n",
       "      <td>2.767</td>\n",
       "      <td>4.173</td>\n",
       "      <td>2.648</td>\n",
       "      <td>6.301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0000066</th>\n",
       "      <td>1.886</td>\n",
       "      <td>-6.084</td>\n",
       "      <td>1.181</td>\n",
       "      <td>0.947</td>\n",
       "      <td>-5.249</td>\n",
       "      <td>1.777</td>\n",
       "      <td>-2.601</td>\n",
       "      <td>-2.823</td>\n",
       "      <td>-3.993</td>\n",
       "      <td>-4.833</td>\n",
       "      <td>...</td>\n",
       "      <td>1.755</td>\n",
       "      <td>1.312</td>\n",
       "      <td>2.178</td>\n",
       "      <td>1.794</td>\n",
       "      <td>1.595</td>\n",
       "      <td>0.597</td>\n",
       "      <td>0.750</td>\n",
       "      <td>1.746</td>\n",
       "      <td>3.141</td>\n",
       "      <td>4.540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0032116</th>\n",
       "      <td>6.093</td>\n",
       "      <td>5.952</td>\n",
       "      <td>5.927</td>\n",
       "      <td>5.501</td>\n",
       "      <td>5.784</td>\n",
       "      <td>5.053</td>\n",
       "      <td>5.563</td>\n",
       "      <td>5.417</td>\n",
       "      <td>5.693</td>\n",
       "      <td>5.021</td>\n",
       "      <td>...</td>\n",
       "      <td>4.469</td>\n",
       "      <td>5.376</td>\n",
       "      <td>4.720</td>\n",
       "      <td>4.396</td>\n",
       "      <td>4.723</td>\n",
       "      <td>4.477</td>\n",
       "      <td>5.014</td>\n",
       "      <td>4.756</td>\n",
       "      <td>5.304</td>\n",
       "      <td>5.340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0033692</th>\n",
       "      <td>5.431</td>\n",
       "      <td>5.504</td>\n",
       "      <td>5.111</td>\n",
       "      <td>5.272</td>\n",
       "      <td>4.984</td>\n",
       "      <td>5.107</td>\n",
       "      <td>5.121</td>\n",
       "      <td>4.817</td>\n",
       "      <td>4.458</td>\n",
       "      <td>4.804</td>\n",
       "      <td>...</td>\n",
       "      <td>5.022</td>\n",
       "      <td>5.377</td>\n",
       "      <td>4.942</td>\n",
       "      <td>4.784</td>\n",
       "      <td>4.984</td>\n",
       "      <td>5.081</td>\n",
       "      <td>4.992</td>\n",
       "      <td>4.798</td>\n",
       "      <td>5.146</td>\n",
       "      <td>4.870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0035542</th>\n",
       "      <td>4.919</td>\n",
       "      <td>4.165</td>\n",
       "      <td>4.908</td>\n",
       "      <td>4.289</td>\n",
       "      <td>4.876</td>\n",
       "      <td>3.714</td>\n",
       "      <td>4.365</td>\n",
       "      <td>3.654</td>\n",
       "      <td>4.777</td>\n",
       "      <td>3.617</td>\n",
       "      <td>...</td>\n",
       "      <td>3.027</td>\n",
       "      <td>3.948</td>\n",
       "      <td>2.701</td>\n",
       "      <td>3.691</td>\n",
       "      <td>3.332</td>\n",
       "      <td>2.330</td>\n",
       "      <td>3.160</td>\n",
       "      <td>3.361</td>\n",
       "      <td>3.546</td>\n",
       "      <td>3.391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0035703</th>\n",
       "      <td>-5.074</td>\n",
       "      <td>-6.084</td>\n",
       "      <td>-4.989</td>\n",
       "      <td>-5.701</td>\n",
       "      <td>-5.249</td>\n",
       "      <td>-6.099</td>\n",
       "      <td>-5.252</td>\n",
       "      <td>-5.473</td>\n",
       "      <td>-3.993</td>\n",
       "      <td>-4.833</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.440</td>\n",
       "      <td>-5.275</td>\n",
       "      <td>-5.085</td>\n",
       "      <td>-4.932</td>\n",
       "      <td>-5.830</td>\n",
       "      <td>-5.223</td>\n",
       "      <td>-5.110</td>\n",
       "      <td>-5.913</td>\n",
       "      <td>-4.146</td>\n",
       "      <td>-5.436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIMAT0035704</th>\n",
       "      <td>-5.074</td>\n",
       "      <td>-6.084</td>\n",
       "      <td>-4.989</td>\n",
       "      <td>-5.701</td>\n",
       "      <td>-5.249</td>\n",
       "      <td>-6.099</td>\n",
       "      <td>-5.252</td>\n",
       "      <td>-5.473</td>\n",
       "      <td>-3.993</td>\n",
       "      <td>-4.833</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.440</td>\n",
       "      <td>-5.275</td>\n",
       "      <td>-5.085</td>\n",
       "      <td>-4.932</td>\n",
       "      <td>-5.830</td>\n",
       "      <td>-5.223</td>\n",
       "      <td>-5.110</td>\n",
       "      <td>-5.913</td>\n",
       "      <td>1.092</td>\n",
       "      <td>-5.436</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2565 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM4158202  GSM4158203  GSM4158204  GSM4158205  GSM4158206  \\\n",
       "ID                                                                         \n",
       "MIMAT0000062       2.840       2.634       3.026       2.769       2.793   \n",
       "MIMAT0000063       3.267       3.881       3.554       3.367       3.403   \n",
       "MIMAT0000064       2.753       2.696       3.104       2.567       3.514   \n",
       "MIMAT0000065       2.615       2.536       2.677       2.441       2.997   \n",
       "MIMAT0000066       1.886      -6.084       1.181       0.947      -5.249   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "MIMAT0032116       6.093       5.952       5.927       5.501       5.784   \n",
       "MIMAT0033692       5.431       5.504       5.111       5.272       4.984   \n",
       "MIMAT0035542       4.919       4.165       4.908       4.289       4.876   \n",
       "MIMAT0035703      -5.074      -6.084      -4.989      -5.701      -5.249   \n",
       "MIMAT0035704      -5.074      -6.084      -4.989      -5.701      -5.249   \n",
       "\n",
       "              GSM4158207  GSM4158208  GSM4158209  GSM4158210  GSM4158211  ...  \\\n",
       "ID                                                                        ...   \n",
       "MIMAT0000062       4.163       3.741       4.383       4.771      -4.833  ...   \n",
       "MIMAT0000063       4.514       4.381       4.681       4.846      -3.565  ...   \n",
       "MIMAT0000064       3.866       3.775       4.007       4.458      -4.833  ...   \n",
       "MIMAT0000065       3.737       3.148       4.080       4.290      -4.833  ...   \n",
       "MIMAT0000066       1.777      -2.601      -2.823      -3.993      -4.833  ...   \n",
       "...                  ...         ...         ...         ...         ...  ...   \n",
       "MIMAT0032116       5.053       5.563       5.417       5.693       5.021  ...   \n",
       "MIMAT0033692       5.107       5.121       4.817       4.458       4.804  ...   \n",
       "MIMAT0035542       3.714       4.365       3.654       4.777       3.617  ...   \n",
       "MIMAT0035703      -6.099      -5.252      -5.473      -3.993      -4.833  ...   \n",
       "MIMAT0035704      -6.099      -5.252      -5.473      -3.993      -4.833  ...   \n",
       "\n",
       "              GSM4158334  GSM4158335  GSM4158336  GSM4158337  GSM4158338  \\\n",
       "ID                                                                         \n",
       "MIMAT0000062       3.863       3.344       4.191       4.145       4.225   \n",
       "MIMAT0000063       4.468       4.054       4.552       4.003       4.497   \n",
       "MIMAT0000064       3.674       3.512       4.011       3.327       4.058   \n",
       "MIMAT0000065       3.232       3.245       3.758       2.931       3.865   \n",
       "MIMAT0000066       1.755       1.312       2.178       1.794       1.595   \n",
       "...                  ...         ...         ...         ...         ...   \n",
       "MIMAT0032116       4.469       5.376       4.720       4.396       4.723   \n",
       "MIMAT0033692       5.022       5.377       4.942       4.784       4.984   \n",
       "MIMAT0035542       3.027       3.948       2.701       3.691       3.332   \n",
       "MIMAT0035703      -5.440      -5.275      -5.085      -4.932      -5.830   \n",
       "MIMAT0035704      -5.440      -5.275      -5.085      -4.932      -5.830   \n",
       "\n",
       "              GSM4158339  GSM4158340  GSM4158341  GSM4158342  GSM4158343  \n",
       "ID                                                                        \n",
       "MIMAT0000062       3.173       3.030       4.526       2.904       6.561  \n",
       "MIMAT0000063       4.079       4.087       4.730       4.321       6.585  \n",
       "MIMAT0000064       3.185       3.097       4.573       3.593       6.395  \n",
       "MIMAT0000065       2.710       2.767       4.173       2.648       6.301  \n",
       "MIMAT0000066       0.597       0.750       1.746       3.141       4.540  \n",
       "...                  ...         ...         ...         ...         ...  \n",
       "MIMAT0032116       4.477       5.014       4.756       5.304       5.340  \n",
       "MIMAT0033692       5.081       4.992       4.798       5.146       4.870  \n",
       "MIMAT0035542       2.330       3.160       3.361       3.546       3.391  \n",
       "MIMAT0035703      -5.223      -5.110      -5.913      -4.146      -5.436  \n",
       "MIMAT0035704      -5.223      -5.110      -5.913       1.092      -5.436  \n",
       "\n",
       "[2565 rows x 142 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['MIMAT0000062', 'MIMAT0000063', 'MIMAT0000064', 'MIMAT0000065', 'MIMAT0000066'], 'miRNA': ['MIMAT0000062', 'MIMAT0000063', 'MIMAT0000064', 'MIMAT0000065', 'MIMAT0000066'], 'miRNA_ID_LIST': ['hsa-let-7a-5p', 'hsa-let-7b-5p', 'hsa-let-7c-5p', 'hsa-let-7d-5p', 'hsa-let-7e-5p']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['MIMAT0000062', 'MIMAT0000063', 'MIMAT0000064', 'MIMAT0000065', 'MIMAT0000066', 'MIMAT0000067', 'MIMAT0000068', 'MIMAT0000069', 'MIMAT0000070', 'MIMAT0000071', 'MIMAT0000072', 'MIMAT0000073', 'MIMAT0000074', 'MIMAT0000075', 'MIMAT0000076', 'MIMAT0000077', 'MIMAT0000078', 'MIMAT0000079', 'MIMAT0000080', 'MIMAT0000081']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['MIMAT0000062', 'MIMAT0000063', 'MIMAT0000064', 'MIMAT0000065', 'MIMAT0000066'], 'miRNA': ['MIMAT0000062', 'MIMAT0000063', 'MIMAT0000064', 'MIMAT0000065', 'MIMAT0000066'], 'miRNA_ID_LIST': ['hsa-let-7a-5p', 'hsa-let-7b-5p', 'hsa-let-7c-5p', 'hsa-let-7d-5p', 'hsa-let-7e-5p']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'miRNA', 'miRNA_ID_LIST'], dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'miRNA', 'miRNA_ID_LIST'], dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Loci-specific differences in blood DNA methylation in HBV-negative populations at risk for hepatocellular carcinoma development - post-diagnostic HCC blood samples\"\n",
      "!Series_summary\t\"Genome wide DNA methylation profiling of blood samples collected from patients after diagnosis with hepatocellular carcinoma (HCC) (cases) vs. blood samples collected from healthy individuals without family history of cancer (controls). The Illumina Infinium 450K Human DNA methylation Beadchip v1.2 was used to obtain DNA methylation profiles across approximately 450,000 CpGs in human samples corresponding to cases (post-diagnostic HCC) and controls. Samples included 24 cases and 24 controls. Cases were matched with controls on gender, age, ethnicity, hepatitis C infection, and diabetes. The presence or absence of HCC in our study was determined based on the AASLD criteria.\"\n",
      "!Series_overall_design\t\"Bisulphite converted DNA from the 42 samples were hybridised to the Illumina Infinium 450k Human Methylation Beadchip v1.2\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM3105174</th>\n",
       "      <th>GSM3105175</th>\n",
       "      <th>GSM3105176</th>\n",
       "      <th>GSM3105177</th>\n",
       "      <th>GSM3105178</th>\n",
       "      <th>GSM3105179</th>\n",
       "      <th>GSM3105180</th>\n",
       "      <th>GSM3105181</th>\n",
       "      <th>GSM3105182</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3105212</th>\n",
       "      <th>GSM3105213</th>\n",
       "      <th>GSM3105214</th>\n",
       "      <th>GSM3105215</th>\n",
       "      <th>GSM3105216</th>\n",
       "      <th>GSM3105217</th>\n",
       "      <th>GSM3105218</th>\n",
       "      <th>GSM3105219</th>\n",
       "      <th>GSM3105220</th>\n",
       "      <th>GSM3105221</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "      <td>case/control: Case - post-diagnostic hepatocel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>...</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                       GSM3105174  \\\n",
       "0  !Sample_characteristics_ch1  case/control: Control - Healthy   \n",
       "1  !Sample_characteristics_ch1            disease state: normal   \n",
       "2  !Sample_characteristics_ch1   sample type: human whole blood   \n",
       "\n",
       "                        GSM3105175                       GSM3105176  \\\n",
       "0  case/control: Control - Healthy  case/control: Control - Healthy   \n",
       "1            disease state: normal            disease state: normal   \n",
       "2   sample type: human whole blood   sample type: human whole blood   \n",
       "\n",
       "                        GSM3105177                       GSM3105178  \\\n",
       "0  case/control: Control - Healthy  case/control: Control - Healthy   \n",
       "1            disease state: normal            disease state: normal   \n",
       "2   sample type: human whole blood   sample type: human whole blood   \n",
       "\n",
       "                        GSM3105179                       GSM3105180  \\\n",
       "0  case/control: Control - Healthy  case/control: Control - Healthy   \n",
       "1            disease state: normal            disease state: normal   \n",
       "2   sample type: human whole blood   sample type: human whole blood   \n",
       "\n",
       "                        GSM3105181                       GSM3105182  ...  \\\n",
       "0  case/control: Control - Healthy  case/control: Control - Healthy  ...   \n",
       "1            disease state: normal            disease state: normal  ...   \n",
       "2   sample type: human whole blood   sample type: human whole blood  ...   \n",
       "\n",
       "                                          GSM3105212  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105213  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105214  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105215  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105216  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105217  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105218  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105219  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105220  \\\n",
       "0  case/control: Case - post-diagnostic hepatocel...   \n",
       "1      disease state: hepatocellular carcinoma (HCC)   \n",
       "2                     sample type: human whole blood   \n",
       "\n",
       "                                          GSM3105221  \n",
       "0  case/control: Case - post-diagnostic hepatocel...  \n",
       "1      disease state: hepatocellular carcinoma (HCC)  \n",
       "2                     sample type: human whole blood  \n",
       "\n",
       "[3 rows x 49 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MemeoryError\n",
    "cohort = accession_num = \"GSE113409\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease state: normal',\n",
       "       'disease state: hepatocellular carcinoma (HCC)'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease state: hepatocellular carcinoma (HCC)':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3105174</th>\n",
       "      <th>GSM3105175</th>\n",
       "      <th>GSM3105176</th>\n",
       "      <th>GSM3105177</th>\n",
       "      <th>GSM3105178</th>\n",
       "      <th>GSM3105179</th>\n",
       "      <th>GSM3105180</th>\n",
       "      <th>GSM3105181</th>\n",
       "      <th>GSM3105182</th>\n",
       "      <th>GSM3105183</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3105212</th>\n",
       "      <th>GSM3105213</th>\n",
       "      <th>GSM3105214</th>\n",
       "      <th>GSM3105215</th>\n",
       "      <th>GSM3105216</th>\n",
       "      <th>GSM3105217</th>\n",
       "      <th>GSM3105218</th>\n",
       "      <th>GSM3105219</th>\n",
       "      <th>GSM3105220</th>\n",
       "      <th>GSM3105221</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hepatitis</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM3105174  GSM3105175  GSM3105176  GSM3105177  GSM3105178  \\\n",
       "Hepatitis           0           0           0           0           0   \n",
       "\n",
       "           GSM3105179  GSM3105180  GSM3105181  GSM3105182  GSM3105183  ...  \\\n",
       "Hepatitis           0           0           0           0           0  ...   \n",
       "\n",
       "           GSM3105212  GSM3105213  GSM3105214  GSM3105215  GSM3105216  \\\n",
       "Hepatitis           1           1           1           1           1   \n",
       "\n",
       "           GSM3105217  GSM3105218  GSM3105219  GSM3105220  GSM3105221  \n",
       "Hepatitis           1           1           1           1           1  \n",
       "\n",
       "[1 rows x 48 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3105174</th>\n",
       "      <th>GSM3105175</th>\n",
       "      <th>GSM3105176</th>\n",
       "      <th>GSM3105177</th>\n",
       "      <th>GSM3105178</th>\n",
       "      <th>GSM3105179</th>\n",
       "      <th>GSM3105180</th>\n",
       "      <th>GSM3105181</th>\n",
       "      <th>GSM3105182</th>\n",
       "      <th>GSM3105183</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3105212</th>\n",
       "      <th>GSM3105213</th>\n",
       "      <th>GSM3105214</th>\n",
       "      <th>GSM3105215</th>\n",
       "      <th>GSM3105216</th>\n",
       "      <th>GSM3105217</th>\n",
       "      <th>GSM3105218</th>\n",
       "      <th>GSM3105219</th>\n",
       "      <th>GSM3105220</th>\n",
       "      <th>GSM3105221</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000029</th>\n",
       "      <td>0.456141</td>\n",
       "      <td>0.545553</td>\n",
       "      <td>0.491754</td>\n",
       "      <td>0.549807</td>\n",
       "      <td>0.512640</td>\n",
       "      <td>0.627964</td>\n",
       "      <td>0.430656</td>\n",
       "      <td>0.571047</td>\n",
       "      <td>0.499600</td>\n",
       "      <td>0.532660</td>\n",
       "      <td>...</td>\n",
       "      <td>0.496885</td>\n",
       "      <td>0.548694</td>\n",
       "      <td>0.511098</td>\n",
       "      <td>0.618431</td>\n",
       "      <td>0.464638</td>\n",
       "      <td>0.378172</td>\n",
       "      <td>0.484464</td>\n",
       "      <td>0.391136</td>\n",
       "      <td>0.546176</td>\n",
       "      <td>0.571302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000108</th>\n",
       "      <td>0.951474</td>\n",
       "      <td>0.952636</td>\n",
       "      <td>0.951960</td>\n",
       "      <td>0.942636</td>\n",
       "      <td>0.963885</td>\n",
       "      <td>0.960909</td>\n",
       "      <td>0.935152</td>\n",
       "      <td>0.961977</td>\n",
       "      <td>0.933759</td>\n",
       "      <td>0.938655</td>\n",
       "      <td>...</td>\n",
       "      <td>0.961619</td>\n",
       "      <td>0.961346</td>\n",
       "      <td>0.964522</td>\n",
       "      <td>0.944376</td>\n",
       "      <td>0.936806</td>\n",
       "      <td>0.913477</td>\n",
       "      <td>0.949904</td>\n",
       "      <td>0.951558</td>\n",
       "      <td>0.954932</td>\n",
       "      <td>0.922968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000109</th>\n",
       "      <td>0.899507</td>\n",
       "      <td>0.896365</td>\n",
       "      <td>0.858829</td>\n",
       "      <td>0.844626</td>\n",
       "      <td>0.797604</td>\n",
       "      <td>0.841048</td>\n",
       "      <td>0.865766</td>\n",
       "      <td>0.871188</td>\n",
       "      <td>0.895762</td>\n",
       "      <td>0.913243</td>\n",
       "      <td>...</td>\n",
       "      <td>0.834972</td>\n",
       "      <td>0.883837</td>\n",
       "      <td>0.878783</td>\n",
       "      <td>0.808218</td>\n",
       "      <td>0.866698</td>\n",
       "      <td>0.851090</td>\n",
       "      <td>0.830295</td>\n",
       "      <td>0.888915</td>\n",
       "      <td>0.869469</td>\n",
       "      <td>0.794161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000165</th>\n",
       "      <td>0.162161</td>\n",
       "      <td>0.140993</td>\n",
       "      <td>0.179489</td>\n",
       "      <td>0.183277</td>\n",
       "      <td>0.182863</td>\n",
       "      <td>0.217246</td>\n",
       "      <td>0.217256</td>\n",
       "      <td>0.236196</td>\n",
       "      <td>0.157192</td>\n",
       "      <td>0.197825</td>\n",
       "      <td>...</td>\n",
       "      <td>0.134914</td>\n",
       "      <td>0.185716</td>\n",
       "      <td>0.194467</td>\n",
       "      <td>0.155407</td>\n",
       "      <td>0.182223</td>\n",
       "      <td>0.471573</td>\n",
       "      <td>0.126177</td>\n",
       "      <td>0.153053</td>\n",
       "      <td>0.167795</td>\n",
       "      <td>0.152402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000236</th>\n",
       "      <td>0.783451</td>\n",
       "      <td>0.719850</td>\n",
       "      <td>0.743973</td>\n",
       "      <td>0.735974</td>\n",
       "      <td>0.790358</td>\n",
       "      <td>0.827679</td>\n",
       "      <td>0.718317</td>\n",
       "      <td>0.772257</td>\n",
       "      <td>0.732836</td>\n",
       "      <td>0.781257</td>\n",
       "      <td>...</td>\n",
       "      <td>0.759885</td>\n",
       "      <td>0.738814</td>\n",
       "      <td>0.751406</td>\n",
       "      <td>0.840779</td>\n",
       "      <td>0.782726</td>\n",
       "      <td>0.740348</td>\n",
       "      <td>0.699439</td>\n",
       "      <td>0.675652</td>\n",
       "      <td>0.716671</td>\n",
       "      <td>0.719669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs9363764</th>\n",
       "      <td>0.884993</td>\n",
       "      <td>0.575390</td>\n",
       "      <td>0.033317</td>\n",
       "      <td>0.575411</td>\n",
       "      <td>0.036038</td>\n",
       "      <td>0.547606</td>\n",
       "      <td>0.039580</td>\n",
       "      <td>0.042940</td>\n",
       "      <td>0.584236</td>\n",
       "      <td>0.550538</td>\n",
       "      <td>...</td>\n",
       "      <td>0.047767</td>\n",
       "      <td>0.582720</td>\n",
       "      <td>0.910069</td>\n",
       "      <td>0.581646</td>\n",
       "      <td>0.943216</td>\n",
       "      <td>0.089709</td>\n",
       "      <td>0.923271</td>\n",
       "      <td>0.935945</td>\n",
       "      <td>0.546927</td>\n",
       "      <td>0.030692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs939290</th>\n",
       "      <td>0.947329</td>\n",
       "      <td>0.947207</td>\n",
       "      <td>0.525745</td>\n",
       "      <td>0.496984</td>\n",
       "      <td>0.030874</td>\n",
       "      <td>0.506133</td>\n",
       "      <td>0.954376</td>\n",
       "      <td>0.951915</td>\n",
       "      <td>0.022362</td>\n",
       "      <td>0.500149</td>\n",
       "      <td>...</td>\n",
       "      <td>0.956040</td>\n",
       "      <td>0.955622</td>\n",
       "      <td>0.505769</td>\n",
       "      <td>0.954314</td>\n",
       "      <td>0.533835</td>\n",
       "      <td>0.949370</td>\n",
       "      <td>0.516095</td>\n",
       "      <td>0.511649</td>\n",
       "      <td>0.943399</td>\n",
       "      <td>0.677327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs951295</th>\n",
       "      <td>0.946336</td>\n",
       "      <td>0.521053</td>\n",
       "      <td>0.525385</td>\n",
       "      <td>0.955580</td>\n",
       "      <td>0.520154</td>\n",
       "      <td>0.050719</td>\n",
       "      <td>0.512903</td>\n",
       "      <td>0.510637</td>\n",
       "      <td>0.525594</td>\n",
       "      <td>0.946099</td>\n",
       "      <td>...</td>\n",
       "      <td>0.509571</td>\n",
       "      <td>0.050995</td>\n",
       "      <td>0.506762</td>\n",
       "      <td>0.956640</td>\n",
       "      <td>0.947660</td>\n",
       "      <td>0.048015</td>\n",
       "      <td>0.053336</td>\n",
       "      <td>0.509812</td>\n",
       "      <td>0.512275</td>\n",
       "      <td>0.050468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs966367</th>\n",
       "      <td>0.032238</td>\n",
       "      <td>0.938334</td>\n",
       "      <td>0.518474</td>\n",
       "      <td>0.951967</td>\n",
       "      <td>0.508393</td>\n",
       "      <td>0.497000</td>\n",
       "      <td>0.927344</td>\n",
       "      <td>0.017595</td>\n",
       "      <td>0.474833</td>\n",
       "      <td>0.948512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.500889</td>\n",
       "      <td>0.021498</td>\n",
       "      <td>0.485564</td>\n",
       "      <td>0.943506</td>\n",
       "      <td>0.465493</td>\n",
       "      <td>0.447545</td>\n",
       "      <td>0.520841</td>\n",
       "      <td>0.480652</td>\n",
       "      <td>0.462363</td>\n",
       "      <td>0.027305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs9839873</th>\n",
       "      <td>0.603442</td>\n",
       "      <td>0.917194</td>\n",
       "      <td>0.908021</td>\n",
       "      <td>0.936910</td>\n",
       "      <td>0.887390</td>\n",
       "      <td>0.910850</td>\n",
       "      <td>0.630803</td>\n",
       "      <td>0.610568</td>\n",
       "      <td>0.862847</td>\n",
       "      <td>0.909776</td>\n",
       "      <td>...</td>\n",
       "      <td>0.893603</td>\n",
       "      <td>0.920107</td>\n",
       "      <td>0.901019</td>\n",
       "      <td>0.083389</td>\n",
       "      <td>0.624759</td>\n",
       "      <td>0.094165</td>\n",
       "      <td>0.633676</td>\n",
       "      <td>0.657978</td>\n",
       "      <td>0.915837</td>\n",
       "      <td>0.681146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>478570 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM3105174  GSM3105175  GSM3105176  GSM3105177  GSM3105178  \\\n",
       "ID                                                                       \n",
       "cg00000029    0.456141    0.545553    0.491754    0.549807    0.512640   \n",
       "cg00000108    0.951474    0.952636    0.951960    0.942636    0.963885   \n",
       "cg00000109    0.899507    0.896365    0.858829    0.844626    0.797604   \n",
       "cg00000165    0.162161    0.140993    0.179489    0.183277    0.182863   \n",
       "cg00000236    0.783451    0.719850    0.743973    0.735974    0.790358   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "rs9363764     0.884993    0.575390    0.033317    0.575411    0.036038   \n",
       "rs939290      0.947329    0.947207    0.525745    0.496984    0.030874   \n",
       "rs951295      0.946336    0.521053    0.525385    0.955580    0.520154   \n",
       "rs966367      0.032238    0.938334    0.518474    0.951967    0.508393   \n",
       "rs9839873     0.603442    0.917194    0.908021    0.936910    0.887390   \n",
       "\n",
       "            GSM3105179  GSM3105180  GSM3105181  GSM3105182  GSM3105183  ...  \\\n",
       "ID                                                                      ...   \n",
       "cg00000029    0.627964    0.430656    0.571047    0.499600    0.532660  ...   \n",
       "cg00000108    0.960909    0.935152    0.961977    0.933759    0.938655  ...   \n",
       "cg00000109    0.841048    0.865766    0.871188    0.895762    0.913243  ...   \n",
       "cg00000165    0.217246    0.217256    0.236196    0.157192    0.197825  ...   \n",
       "cg00000236    0.827679    0.718317    0.772257    0.732836    0.781257  ...   \n",
       "...                ...         ...         ...         ...         ...  ...   \n",
       "rs9363764     0.547606    0.039580    0.042940    0.584236    0.550538  ...   \n",
       "rs939290      0.506133    0.954376    0.951915    0.022362    0.500149  ...   \n",
       "rs951295      0.050719    0.512903    0.510637    0.525594    0.946099  ...   \n",
       "rs966367      0.497000    0.927344    0.017595    0.474833    0.948512  ...   \n",
       "rs9839873     0.910850    0.630803    0.610568    0.862847    0.909776  ...   \n",
       "\n",
       "            GSM3105212  GSM3105213  GSM3105214  GSM3105215  GSM3105216  \\\n",
       "ID                                                                       \n",
       "cg00000029    0.496885    0.548694    0.511098    0.618431    0.464638   \n",
       "cg00000108    0.961619    0.961346    0.964522    0.944376    0.936806   \n",
       "cg00000109    0.834972    0.883837    0.878783    0.808218    0.866698   \n",
       "cg00000165    0.134914    0.185716    0.194467    0.155407    0.182223   \n",
       "cg00000236    0.759885    0.738814    0.751406    0.840779    0.782726   \n",
       "...                ...         ...         ...         ...         ...   \n",
       "rs9363764     0.047767    0.582720    0.910069    0.581646    0.943216   \n",
       "rs939290      0.956040    0.955622    0.505769    0.954314    0.533835   \n",
       "rs951295      0.509571    0.050995    0.506762    0.956640    0.947660   \n",
       "rs966367      0.500889    0.021498    0.485564    0.943506    0.465493   \n",
       "rs9839873     0.893603    0.920107    0.901019    0.083389    0.624759   \n",
       "\n",
       "            GSM3105217  GSM3105218  GSM3105219  GSM3105220  GSM3105221  \n",
       "ID                                                                      \n",
       "cg00000029    0.378172    0.484464    0.391136    0.546176    0.571302  \n",
       "cg00000108    0.913477    0.949904    0.951558    0.954932    0.922968  \n",
       "cg00000109    0.851090    0.830295    0.888915    0.869469    0.794161  \n",
       "cg00000165    0.471573    0.126177    0.153053    0.167795    0.152402  \n",
       "cg00000236    0.740348    0.699439    0.675652    0.716671    0.719669  \n",
       "...                ...         ...         ...         ...         ...  \n",
       "rs9363764     0.089709    0.923271    0.935945    0.546927    0.030692  \n",
       "rs939290      0.949370    0.516095    0.511649    0.943399    0.677327  \n",
       "rs951295      0.048015    0.053336    0.509812    0.512275    0.050468  \n",
       "rs966367      0.447545    0.520841    0.480652    0.462363    0.027305  \n",
       "rs9839873     0.094165    0.633676    0.657978    0.915837    0.681146  \n",
       "\n",
       "[478570 rows x 48 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[19], line 4\u001b[0m\n\u001b[0;32m      1\u001b[0m requires_gene_mapping \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m requires_gene_mapping:\n\u001b[1;32m----> 4\u001b[0m     gene_annotation \u001b[38;5;241m=\u001b[39m \u001b[43mget_gene_annotation\u001b[49m\u001b[43m(\u001b[49m\u001b[43msoft_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      5\u001b[0m     gene_annotation_summary \u001b[38;5;241m=\u001b[39m preview_df(gene_annotation)\n\u001b[0;32m      6\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene_annotation_summary)\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:128\u001b[0m, in \u001b[0;36mget_gene_annotation\u001b[1;34m(file_path, prefixes)\u001b[0m\n\u001b[0;32m    126\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_gene_annotation\u001b[39m(file_path, prefixes\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m^\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m!\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    127\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Extract from a SOFT file the gene annotation data\"\"\"\u001b[39;00m\n\u001b[1;32m--> 128\u001b[0m     gene_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_content_by_prefix\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefixes_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mprefixes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munselect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfile\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    129\u001b[0m \u001b[43m                                             \u001b[49m\u001b[43mreturn_df_a\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    130\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m gene_metadata[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:106\u001b[0m, in \u001b[0;36mfilter_content_by_prefix\u001b[1;34m(source, prefixes_a, prefixes_b, unselect, source_type, return_df_a, return_df_b)\u001b[0m\n\u001b[0;32m    104\u001b[0m filtered_content_a \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(filtered_lines_a)\n\u001b[0;32m    105\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m return_df_a:\n\u001b[1;32m--> 106\u001b[0m     filtered_content_a \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfiltered_content_a\u001b[49m\u001b[43m)\u001b[49m, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, low_memory\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, on_bad_lines\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mskip\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m    107\u001b[0m filtered_content_b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    108\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filtered_lines_b:\n",
      "\u001b[1;31mMemoryError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Identification of a serum 48-lncRNA signature as diagnostic marker for hepatocellular carcinoma and liver cirrhosis\"\n",
      "!Series_summary\t\"In cancer management, early and accurate diagnosis of hepatocellular carcinoma (HCC) is important for enhancing survival rate of patients. Currently, serum alpha-fetoprotein (AFP) is the only one biomarker for detection of HCC. However, serum AFP is not satisfactory for diagnosis of HCC due to its low accuracy (about 60-70%). In this study, we collected 109 serum samples (discovery set) from healthy control (HC) and patients with chronic hepatitis B (CHB), liver cirrhosis (LC) and HCC, and analyzed them with custom lncRNA microarray. Profiling analysis shows 181 differentially expressed lncRNAs between HCs and patients with CHB, LC and HCC. Then a 48-lncRNA diagnostic signature was identified with 100% predictive accuracy for all subjects in the discovery set. This diagnostic signature was verified with a cross-validation analysis in the discovery set. To further corroborate the signature, we gathered another 66 serum samples (validation set) and also analyzed them with microarray. The result indicates that the same signature has similar diagnostic accuracy for HC (100%), CHB (73%), LC (88%) and HCC (95%), implying a reproducible diagnostic biomarker for HCC. Receiver operating characteristic (ROC) analysis exhibits that this signature has significantly higher diagnostic accuracy for HCC and non-cancerous subjects (area under curve [AUC]: 0.994) than AFP (AUC: 0.773) in the discovery set and this was also verified in the validation set (0.964 vs 0.792). More importantly, the signature detected small HCC (<3cm) with 100% (13/13) accuracy while AFP with only 61.5% (8/13). Altogether, this study demonstrates that the serum 48-lncRNA signature is not only a powerful and sensitive biomarker for diagnosis of HCC but also a potential biomarker for LC.\"\n",
      "!Series_summary\t\"\"\n",
      "!Series_summary\t\"*****************************************************************\"\n",
      "!Series_summary\t\"Submitter declares these data are subject to patent number ZL 2016 1 0397094.\"\n",
      "!Series_summary\t\"*****************************************************************\"\n",
      "!Series_overall_design\t\"In this study, we aimed to profile serum lncRNA expression and identify a signature for diagnosis of HCC. Firstly, we collected 109 serum samples (discovery set) from healthy control (HC) and patients with chronic hepatitis B (CHB), liver cirrhosis (LC) and HCC, and analyzed them with custom lncRNA microarray. To further corroborate the signature, we gathered another 66 serum samples (validation set) and also analyzed them with microarray.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2068378</th>\n",
       "      <th>GSM2068379</th>\n",
       "      <th>GSM2068380</th>\n",
       "      <th>GSM2068381</th>\n",
       "      <th>GSM2068382</th>\n",
       "      <th>GSM2068383</th>\n",
       "      <th>GSM2068384</th>\n",
       "      <th>GSM2068385</th>\n",
       "      <th>GSM2068386</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2068543</th>\n",
       "      <th>GSM2068544</th>\n",
       "      <th>GSM2068545</th>\n",
       "      <th>GSM2068546</th>\n",
       "      <th>GSM2068547</th>\n",
       "      <th>GSM2068548</th>\n",
       "      <th>GSM2068549</th>\n",
       "      <th>GSM2068550</th>\n",
       "      <th>GSM2068551</th>\n",
       "      <th>GSM2068552</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "      <td>tissue: serum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>disease state: healthy control</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "      <td>disease state: hepatocellular carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>data set: discovery</td>\n",
       "      <td>...</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "      <td>data set: validation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 176 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                      GSM2068378  \\\n",
       "0  !Sample_characteristics_ch1                   tissue: serum   \n",
       "1  !Sample_characteristics_ch1  disease state: healthy control   \n",
       "2  !Sample_characteristics_ch1             data set: discovery   \n",
       "\n",
       "                       GSM2068379                      GSM2068380  \\\n",
       "0                   tissue: serum                   tissue: serum   \n",
       "1  disease state: healthy control  disease state: healthy control   \n",
       "2             data set: discovery             data set: discovery   \n",
       "\n",
       "                       GSM2068381                      GSM2068382  \\\n",
       "0                   tissue: serum                   tissue: serum   \n",
       "1  disease state: healthy control  disease state: healthy control   \n",
       "2             data set: discovery             data set: discovery   \n",
       "\n",
       "                       GSM2068383                      GSM2068384  \\\n",
       "0                   tissue: serum                   tissue: serum   \n",
       "1  disease state: healthy control  disease state: healthy control   \n",
       "2             data set: discovery             data set: discovery   \n",
       "\n",
       "                       GSM2068385                      GSM2068386  ...  \\\n",
       "0                   tissue: serum                   tissue: serum  ...   \n",
       "1  disease state: healthy control  disease state: healthy control  ...   \n",
       "2             data set: discovery             data set: discovery  ...   \n",
       "\n",
       "                                GSM2068543  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068544  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068545  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068546  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068547  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068548  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068549  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068550  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068551  \\\n",
       "0                            tissue: serum   \n",
       "1  disease state: hepatocellular carcinoma   \n",
       "2                     data set: validation   \n",
       "\n",
       "                                GSM2068552  \n",
       "0                            tissue: serum  \n",
       "1  disease state: hepatocellular carcinoma  \n",
       "2                     data set: validation  \n",
       "\n",
       "[3 rows x 176 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE78160\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease state: healthy control',\n",
       "       'disease state: chronic hepatitis B',\n",
       "       'disease state: liver cirrhosis',\n",
       "       'disease state: hepatocellular carcinoma'], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease state: chronic hepatitis B':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2068378</th>\n",
       "      <th>GSM2068379</th>\n",
       "      <th>GSM2068380</th>\n",
       "      <th>GSM2068381</th>\n",
       "      <th>GSM2068382</th>\n",
       "      <th>GSM2068383</th>\n",
       "      <th>GSM2068384</th>\n",
       "      <th>GSM2068385</th>\n",
       "      <th>GSM2068386</th>\n",
       "      <th>GSM2068387</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2068543</th>\n",
       "      <th>GSM2068544</th>\n",
       "      <th>GSM2068545</th>\n",
       "      <th>GSM2068546</th>\n",
       "      <th>GSM2068547</th>\n",
       "      <th>GSM2068548</th>\n",
       "      <th>GSM2068549</th>\n",
       "      <th>GSM2068550</th>\n",
       "      <th>GSM2068551</th>\n",
       "      <th>GSM2068552</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hepatitis</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 175 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM2068378  GSM2068379  GSM2068380  GSM2068381  GSM2068382  \\\n",
       "Hepatitis           0           0           0           0           0   \n",
       "\n",
       "           GSM2068383  GSM2068384  GSM2068385  GSM2068386  GSM2068387  ...  \\\n",
       "Hepatitis           0           0           0           0           0  ...   \n",
       "\n",
       "           GSM2068543  GSM2068544  GSM2068545  GSM2068546  GSM2068547  \\\n",
       "Hepatitis           0           0           0           0           0   \n",
       "\n",
       "           GSM2068548  GSM2068549  GSM2068550  GSM2068551  GSM2068552  \n",
       "Hepatitis           0           0           0           0           0  \n",
       "\n",
       "[1 rows x 175 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2068378</th>\n",
       "      <th>GSM2068379</th>\n",
       "      <th>GSM2068380</th>\n",
       "      <th>GSM2068381</th>\n",
       "      <th>GSM2068382</th>\n",
       "      <th>GSM2068383</th>\n",
       "      <th>GSM2068384</th>\n",
       "      <th>GSM2068385</th>\n",
       "      <th>GSM2068386</th>\n",
       "      <th>GSM2068387</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2068543</th>\n",
       "      <th>GSM2068544</th>\n",
       "      <th>GSM2068545</th>\n",
       "      <th>GSM2068546</th>\n",
       "      <th>GSM2068547</th>\n",
       "      <th>GSM2068548</th>\n",
       "      <th>GSM2068549</th>\n",
       "      <th>GSM2068550</th>\n",
       "      <th>GSM2068551</th>\n",
       "      <th>GSM2068552</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>12.123889</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.890497</td>\n",
       "      <td>9.298094</td>\n",
       "      <td>10.515535</td>\n",
       "      <td>10.303984</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>12.128642</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.474615</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2516</th>\n",
       "      <td>14.147935</td>\n",
       "      <td>14.201631</td>\n",
       "      <td>13.710442</td>\n",
       "      <td>14.055458</td>\n",
       "      <td>13.706570</td>\n",
       "      <td>12.381570</td>\n",
       "      <td>14.072761</td>\n",
       "      <td>12.585021</td>\n",
       "      <td>13.918598</td>\n",
       "      <td>14.086638</td>\n",
       "      <td>...</td>\n",
       "      <td>15.719566</td>\n",
       "      <td>13.831713</td>\n",
       "      <td>14.923421</td>\n",
       "      <td>14.254049</td>\n",
       "      <td>14.505217</td>\n",
       "      <td>12.853661</td>\n",
       "      <td>13.391889</td>\n",
       "      <td>15.424096</td>\n",
       "      <td>14.041115</td>\n",
       "      <td>14.297063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2517</th>\n",
       "      <td>14.293053</td>\n",
       "      <td>15.253498</td>\n",
       "      <td>14.591874</td>\n",
       "      <td>15.000236</td>\n",
       "      <td>15.315987</td>\n",
       "      <td>15.531363</td>\n",
       "      <td>15.179144</td>\n",
       "      <td>14.094255</td>\n",
       "      <td>15.180399</td>\n",
       "      <td>15.388774</td>\n",
       "      <td>...</td>\n",
       "      <td>14.262666</td>\n",
       "      <td>12.777848</td>\n",
       "      <td>13.929949</td>\n",
       "      <td>13.053648</td>\n",
       "      <td>13.675200</td>\n",
       "      <td>13.590098</td>\n",
       "      <td>14.167543</td>\n",
       "      <td>15.472563</td>\n",
       "      <td>14.227284</td>\n",
       "      <td>14.443754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2518</th>\n",
       "      <td>14.651719</td>\n",
       "      <td>10.878482</td>\n",
       "      <td>12.772778</td>\n",
       "      <td>12.068672</td>\n",
       "      <td>12.434913</td>\n",
       "      <td>13.155134</td>\n",
       "      <td>12.168948</td>\n",
       "      <td>11.266066</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>12.914463</td>\n",
       "      <td>...</td>\n",
       "      <td>14.380008</td>\n",
       "      <td>12.462397</td>\n",
       "      <td>13.909031</td>\n",
       "      <td>13.137195</td>\n",
       "      <td>13.509931</td>\n",
       "      <td>14.440875</td>\n",
       "      <td>15.480936</td>\n",
       "      <td>15.092393</td>\n",
       "      <td>13.922251</td>\n",
       "      <td>13.806244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2519</th>\n",
       "      <td>14.704295</td>\n",
       "      <td>15.118519</td>\n",
       "      <td>15.235386</td>\n",
       "      <td>15.152312</td>\n",
       "      <td>15.855261</td>\n",
       "      <td>15.921530</td>\n",
       "      <td>15.680007</td>\n",
       "      <td>14.768685</td>\n",
       "      <td>15.588718</td>\n",
       "      <td>15.937210</td>\n",
       "      <td>...</td>\n",
       "      <td>15.719566</td>\n",
       "      <td>15.964852</td>\n",
       "      <td>15.847416</td>\n",
       "      <td>15.964852</td>\n",
       "      <td>15.847416</td>\n",
       "      <td>16.073444</td>\n",
       "      <td>15.964852</td>\n",
       "      <td>15.964852</td>\n",
       "      <td>16.073444</td>\n",
       "      <td>15.569119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2520</th>\n",
       "      <td>14.997646</td>\n",
       "      <td>14.633403</td>\n",
       "      <td>15.144705</td>\n",
       "      <td>15.118480</td>\n",
       "      <td>15.464669</td>\n",
       "      <td>15.348168</td>\n",
       "      <td>15.325703</td>\n",
       "      <td>14.039647</td>\n",
       "      <td>15.386186</td>\n",
       "      <td>15.377220</td>\n",
       "      <td>...</td>\n",
       "      <td>15.719566</td>\n",
       "      <td>14.475181</td>\n",
       "      <td>15.596095</td>\n",
       "      <td>15.015872</td>\n",
       "      <td>14.560201</td>\n",
       "      <td>15.306305</td>\n",
       "      <td>15.964852</td>\n",
       "      <td>15.964852</td>\n",
       "      <td>15.660260</td>\n",
       "      <td>15.797198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2520 rows × 175 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GSM2068378  GSM2068379  GSM2068380  GSM2068381  GSM2068382  GSM2068383  \\\n",
       "ID                                                                             \n",
       "1       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "2       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "3       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "4       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "5       8.643856   10.890497    9.298094   10.515535   10.303984    8.643856   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "2516   14.147935   14.201631   13.710442   14.055458   13.706570   12.381570   \n",
       "2517   14.293053   15.253498   14.591874   15.000236   15.315987   15.531363   \n",
       "2518   14.651719   10.878482   12.772778   12.068672   12.434913   13.155134   \n",
       "2519   14.704295   15.118519   15.235386   15.152312   15.855261   15.921530   \n",
       "2520   14.997646   14.633403   15.144705   15.118480   15.464669   15.348168   \n",
       "\n",
       "      GSM2068384  GSM2068385  GSM2068386  GSM2068387  ...  GSM2068543  \\\n",
       "ID                                                    ...               \n",
       "1       8.643856    8.643856    8.643856    8.643856  ...    8.643856   \n",
       "2       8.643856    8.643856    8.643856    8.643856  ...    8.643856   \n",
       "3       8.643856    8.643856    8.643856    8.643856  ...    8.643856   \n",
       "4       8.643856    8.643856    8.643856    8.643856  ...    8.643856   \n",
       "5       8.643856    8.643856    8.643856    8.643856  ...   12.128642   \n",
       "...          ...         ...         ...         ...  ...         ...   \n",
       "2516   14.072761   12.585021   13.918598   14.086638  ...   15.719566   \n",
       "2517   15.179144   14.094255   15.180399   15.388774  ...   14.262666   \n",
       "2518   12.168948   11.266066    8.643856   12.914463  ...   14.380008   \n",
       "2519   15.680007   14.768685   15.588718   15.937210  ...   15.719566   \n",
       "2520   15.325703   14.039647   15.386186   15.377220  ...   15.719566   \n",
       "\n",
       "      GSM2068544  GSM2068545  GSM2068546  GSM2068547  GSM2068548  GSM2068549  \\\n",
       "ID                                                                             \n",
       "1       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "2       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "3       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "4       8.643856    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "5       8.643856    8.643856    8.643856   10.474615    8.643856    8.643856   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "2516   13.831713   14.923421   14.254049   14.505217   12.853661   13.391889   \n",
       "2517   12.777848   13.929949   13.053648   13.675200   13.590098   14.167543   \n",
       "2518   12.462397   13.909031   13.137195   13.509931   14.440875   15.480936   \n",
       "2519   15.964852   15.847416   15.964852   15.847416   16.073444   15.964852   \n",
       "2520   14.475181   15.596095   15.015872   14.560201   15.306305   15.964852   \n",
       "\n",
       "      GSM2068550  GSM2068551  GSM2068552  \n",
       "ID                                        \n",
       "1       8.643856    8.643856    8.643856  \n",
       "2       8.643856    8.643856    8.643856  \n",
       "3       8.643856    8.643856    8.643856  \n",
       "4       8.643856   12.123889    8.643856  \n",
       "5       8.643856    8.643856    8.643856  \n",
       "...          ...         ...         ...  \n",
       "2516   15.424096   14.041115   14.297063  \n",
       "2517   15.472563   14.227284   14.443754  \n",
       "2518   15.092393   13.922251   13.806244  \n",
       "2519   15.964852   16.073444   15.569119  \n",
       "2520   15.964852   15.660260   15.797198  \n",
       "\n",
       "[2520 rows x 175 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1', '2', '3', '4', '5'], 'SPOT_ID': ['UBC', 'NEIF3D', 'GRM4', 'GAPDH', 'NC-1'], 'SEQUENCE': ['TCGAGCCCAGTGACACCATCGAGAACGTCAAAGCAAAGATCCAGGACAAGGAAGG', 'ACTACGACAAAGCCTTTGACCGCATCACCACGAGGAGTGAGAAGCCACTGCGGAGCAT', 'ATCCTCTTTGTCCTCAGCTCCTCCTGCTTTCTTGGGTCCCACCAGTGTCACTTTTCTGCC', 'AGTGGTGGACCTGACCTGCCGTCTAGAAAAACCTGCCAAATATGATGACATCAAGAAG', 'TCAGAGTAGACTCGACTGCGATGTACTCAGCACAGCAACGTCGTAGGACAGTCGATCTTG']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1', '2', '3', '4', '5'], 'SPOT_ID': ['UBC', 'NEIF3D', 'GRM4', 'GAPDH', 'NC-1'], 'SEQUENCE': ['TCGAGCCCAGTGACACCATCGAGAACGTCAAAGCAAAGATCCAGGACAAGGAAGG', 'ACTACGACAAAGCCTTTGACCGCATCACCACGAGGAGTGAGAAGCCACTGCGGAGCAT', 'ATCCTCTTTGTCCTCAGCTCCTCCTGCTTTCTTGGGTCCCACCAGTGTCACTTTTCTGCC', 'AGTGGTGGACCTGACCTGCCGTCTAGAAAAACCTGCCAAATATGATGACATCAAGAAG', 'TCAGAGTAGACTCGACTGCGATGTACTCAGCACAGCAACGTCGTAGGACAGTCGATCTTG']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'SPOT_ID', 'SEQUENCE'], dtype='object')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'SPOT_ID'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM2068378</th>\n",
       "      <th>GSM2068379</th>\n",
       "      <th>GSM2068380</th>\n",
       "      <th>GSM2068381</th>\n",
       "      <th>GSM2068382</th>\n",
       "      <th>GSM2068383</th>\n",
       "      <th>GSM2068384</th>\n",
       "      <th>GSM2068385</th>\n",
       "      <th>GSM2068386</th>\n",
       "      <th>GSM2068387</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2068543</th>\n",
       "      <th>GSM2068544</th>\n",
       "      <th>GSM2068545</th>\n",
       "      <th>GSM2068546</th>\n",
       "      <th>GSM2068547</th>\n",
       "      <th>GSM2068548</th>\n",
       "      <th>GSM2068549</th>\n",
       "      <th>GSM2068550</th>\n",
       "      <th>GSM2068551</th>\n",
       "      <th>GSM2068552</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AIR</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>11.372260</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COPG2IT1</th>\n",
       "      <td>9.603117</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DGCR5</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.847305</td>\n",
       "      <td>10.266793</td>\n",
       "      <td>11.977783</td>\n",
       "      <td>11.383699</td>\n",
       "      <td>10.626553</td>\n",
       "      <td>10.721836</td>\n",
       "      <td>9.398005</td>\n",
       "      <td>11.262412</td>\n",
       "      <td>11.078104</td>\n",
       "      <td>...</td>\n",
       "      <td>12.946249</td>\n",
       "      <td>8.952611</td>\n",
       "      <td>10.381489</td>\n",
       "      <td>10.757499</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.477524</td>\n",
       "      <td>11.104535</td>\n",
       "      <td>12.809209</td>\n",
       "      <td>11.095331</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DISC2</th>\n",
       "      <td>11.094534</td>\n",
       "      <td>9.750400</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.825720</td>\n",
       "      <td>10.434356</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.771369</td>\n",
       "      <td>9.763246</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.355665</td>\n",
       "      <td>...</td>\n",
       "      <td>12.394881</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.579984</td>\n",
       "      <td>9.680585</td>\n",
       "      <td>10.862449</td>\n",
       "      <td>9.290551</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.758394</td>\n",
       "      <td>10.532171</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DIO3OS</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.959334</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMX2OS</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAPDH</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.929325</td>\n",
       "      <td>8.686277</td>\n",
       "      <td>9.094759</td>\n",
       "      <td>8.915263</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>9.856177</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.783056</td>\n",
       "      <td>8.770832</td>\n",
       "      <td>9.167772</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.033588</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.304445</td>\n",
       "      <td>10.007780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAS5</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRM4</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.938785</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.057079</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.751193</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.762620</td>\n",
       "      <td>...</td>\n",
       "      <td>9.739373</td>\n",
       "      <td>8.751991</td>\n",
       "      <td>9.431670</td>\n",
       "      <td>8.738022</td>\n",
       "      <td>8.869630</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.516300</td>\n",
       "      <td>8.655169</td>\n",
       "      <td>11.837607</td>\n",
       "      <td>11.610034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H19</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.256151</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.097203</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.128091</td>\n",
       "      <td>9.653112</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.500107</td>\n",
       "      <td>9.120463</td>\n",
       "      <td>...</td>\n",
       "      <td>11.145225</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.807397</td>\n",
       "      <td>9.956303</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HAR1A</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.138550</td>\n",
       "      <td>9.378628</td>\n",
       "      <td>9.059424</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.165951</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.955595</td>\n",
       "      <td>10.487835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HAR1B</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.168098</td>\n",
       "      <td>9.302220</td>\n",
       "      <td>9.765537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOTAIR</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.070974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOTAIRM1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.867965</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.833051</td>\n",
       "      <td>9.139042</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.949605</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOTTIP</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HULC</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IPW</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JPX</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.416837</td>\n",
       "      <td>10.760934</td>\n",
       "      <td>11.395186</td>\n",
       "      <td>12.195077</td>\n",
       "      <td>11.428226</td>\n",
       "      <td>11.669276</td>\n",
       "      <td>11.293093</td>\n",
       "      <td>12.224064</td>\n",
       "      <td>11.606849</td>\n",
       "      <td>...</td>\n",
       "      <td>12.051680</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.164085</td>\n",
       "      <td>9.320276</td>\n",
       "      <td>10.462688</td>\n",
       "      <td>11.692138</td>\n",
       "      <td>11.618579</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRASP1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KCNQ1OT1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEG3</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>13.022121</td>\n",
       "      <td>11.877640</td>\n",
       "      <td>12.842724</td>\n",
       "      <td>12.358173</td>\n",
       "      <td>12.021758</td>\n",
       "      <td>11.923220</td>\n",
       "      <td>11.031953</td>\n",
       "      <td>12.708179</td>\n",
       "      <td>11.860068</td>\n",
       "      <td>...</td>\n",
       "      <td>12.957414</td>\n",
       "      <td>9.588119</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.265769</td>\n",
       "      <td>8.966986</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.799453</td>\n",
       "      <td>12.503942</td>\n",
       "      <td>13.704860</td>\n",
       "      <td>13.976352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MESTIT1</th>\n",
       "      <td>9.591102</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.364609</td>\n",
       "      <td>11.280995</td>\n",
       "      <td>9.127304</td>\n",
       "      <td>9.820855</td>\n",
       "      <td>10.897501</td>\n",
       "      <td>10.793347</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.548344</td>\n",
       "      <td>...</td>\n",
       "      <td>12.538985</td>\n",
       "      <td>10.952623</td>\n",
       "      <td>10.424415</td>\n",
       "      <td>11.111416</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.120307</td>\n",
       "      <td>11.617112</td>\n",
       "      <td>13.244554</td>\n",
       "      <td>13.167716</td>\n",
       "      <td>13.273113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MALAT1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEAT1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NRON</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NTT</th>\n",
       "      <td>10.959235</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.185743</td>\n",
       "      <td>10.335539</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCGEM1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRINS</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRNCR1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTENP1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNHG1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNHG3</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.927197</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.153955</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.261016</td>\n",
       "      <td>...</td>\n",
       "      <td>11.854897</td>\n",
       "      <td>9.494135</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.601943</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNHG4</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNHG5</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNHG6</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPRY4-IT1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TUG1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.537742</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.584173</td>\n",
       "      <td>8.892180</td>\n",
       "      <td>9.199929</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.588904</td>\n",
       "      <td>...</td>\n",
       "      <td>9.078499</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.150089</td>\n",
       "      <td>10.231673</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBC</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.652308</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.775695</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.763658</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCA1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WT1-AS</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XIST</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.070961</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.807389</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFAS1</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.942281</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MASCRNA</th>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44 rows × 175 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM2068378  GSM2068379  GSM2068380  GSM2068381  GSM2068382  \\\n",
       "7SK          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "AIR          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "COPG2IT1     9.603117    8.643856    8.643856    8.643856    8.643856   \n",
       "DGCR5        8.643856   11.847305   10.266793   11.977783   11.383699   \n",
       "DISC2       11.094534    9.750400    8.643856   10.825720   10.434356   \n",
       "DIO3OS       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "EMX2OS       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "GAPDH        8.643856    8.929325    8.686277    9.094759    8.915263   \n",
       "GAS5         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "GRM4         8.643856    8.938785    8.643856    9.057079    8.643856   \n",
       "H19          8.643856    9.256151    8.643856   10.097203    8.643856   \n",
       "HAR1A        8.643856   10.138550    9.378628    9.059424    8.643856   \n",
       "HAR1B        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HOTAIR       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HOTAIRM1     8.643856   10.867965    8.643856   10.833051    9.139042   \n",
       "HOTTIP       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HULC         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "IPW          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "JPX          8.643856   11.416837   10.760934   11.395186   12.195077   \n",
       "KRASP1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "KCNQ1OT1     8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "MEG3         8.643856   13.022121   11.877640   12.842724   12.358173   \n",
       "MESTIT1      9.591102    8.643856   10.364609   11.280995    9.127304   \n",
       "MALAT1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "NEAT1        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "NRON         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "NTT         10.959235    8.643856    8.643856    8.643856    8.643856   \n",
       "PCGEM1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PRINS        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PRNCR1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PTENP1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG1        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG3        8.643856   10.927197    8.643856   11.153955    8.643856   \n",
       "SNHG4        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG5        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG6        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SPRY4-IT1    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "TUG1         8.643856   11.537742    8.643856   11.584173    8.892180   \n",
       "UBC          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "UCA1         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "WT1-AS       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "XIST         8.643856   10.070961    8.643856    8.807389    8.643856   \n",
       "ZFAS1        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "MASCRNA      8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "\n",
       "           GSM2068383  GSM2068384  GSM2068385  GSM2068386  GSM2068387  ...  \\\n",
       "7SK          8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "AIR          8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "COPG2IT1     8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "DGCR5       10.626553   10.721836    9.398005   11.262412   11.078104  ...   \n",
       "DISC2        8.643856   10.771369    9.763246    8.643856   10.355665  ...   \n",
       "DIO3OS       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "EMX2OS       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "GAPDH        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "GAS5         8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "GRM4         8.643856    8.751193    8.643856    8.643856    8.762620  ...   \n",
       "H19         10.128091    9.653112    8.643856    9.500107    9.120463  ...   \n",
       "HAR1A        8.643856    8.643856    8.643856    9.165951    8.643856  ...   \n",
       "HAR1B        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "HOTAIR       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "HOTAIRM1     8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "HOTTIP       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "HULC         8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "IPW          8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "JPX         11.428226   11.669276   11.293093   12.224064   11.606849  ...   \n",
       "KRASP1       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "KCNQ1OT1     8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "MEG3        12.021758   11.923220   11.031953   12.708179   11.860068  ...   \n",
       "MESTIT1      9.820855   10.897501   10.793347    8.643856   11.548344  ...   \n",
       "MALAT1       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "NEAT1        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "NRON         8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "NTT          8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "PCGEM1       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "PRINS        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "PRNCR1       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "PTENP1       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "SNHG1        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "SNHG3        8.643856    8.643856    8.643856    8.643856    9.261016  ...   \n",
       "SNHG4        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "SNHG5        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "SNHG6        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "SPRY4-IT1    8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "TUG1         9.199929    8.643856    8.643856    8.643856    9.588904  ...   \n",
       "UBC          8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "UCA1         8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "WT1-AS       8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "XIST         8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "ZFAS1        8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "MASCRNA      8.643856    8.643856    8.643856    8.643856    8.643856  ...   \n",
       "\n",
       "           GSM2068543  GSM2068544  GSM2068545  GSM2068546  GSM2068547  \\\n",
       "7SK          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "AIR         11.372260    8.643856    8.643856    8.643856    8.643856   \n",
       "COPG2IT1     8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "DGCR5       12.946249    8.952611   10.381489   10.757499    8.643856   \n",
       "DISC2       12.394881    8.643856   11.579984    9.680585   10.862449   \n",
       "DIO3OS       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "EMX2OS       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "GAPDH        9.856177    8.643856    9.783056    8.770832    9.167772   \n",
       "GAS5         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "GRM4         9.739373    8.751991    9.431670    8.738022    8.869630   \n",
       "H19         11.145225    8.643856    8.643856    8.643856    8.643856   \n",
       "HAR1A        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HAR1B        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HOTAIR       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HOTAIRM1     8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HOTTIP       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "HULC         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "IPW          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "JPX         12.051680    8.643856    8.643856   10.164085    9.320276   \n",
       "KRASP1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "KCNQ1OT1     8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "MEG3        12.957414    9.588119    8.643856   10.265769    8.966986   \n",
       "MESTIT1     12.538985   10.952623   10.424415   11.111416    8.643856   \n",
       "MALAT1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "NEAT1        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "NRON         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "NTT          8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PCGEM1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PRINS        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PRNCR1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "PTENP1       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG1        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG3       11.854897    9.494135    8.643856    8.643856    8.643856   \n",
       "SNHG4        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG5        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SNHG6        8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "SPRY4-IT1    8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "TUG1         9.078499    8.643856    8.643856    8.643856    8.643856   \n",
       "UBC          8.652308    8.643856    8.775695    8.643856    8.763658   \n",
       "UCA1         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "WT1-AS       8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "XIST         8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "ZFAS1        8.643856    8.643856    8.643856    8.643856    9.942281   \n",
       "MASCRNA      8.643856    8.643856    8.643856    8.643856    8.643856   \n",
       "\n",
       "           GSM2068548  GSM2068549  GSM2068550  GSM2068551  GSM2068552  \n",
       "7SK          8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "AIR          8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "COPG2IT1     8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "DGCR5        9.477524   11.104535   12.809209   11.095331    8.643856  \n",
       "DISC2        9.290551    8.643856   10.758394   10.532171    8.643856  \n",
       "DIO3OS       8.643856    8.643856    8.959334    8.643856    8.643856  \n",
       "EMX2OS       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "GAPDH        8.643856    9.033588    8.643856   10.304445   10.007780  \n",
       "GAS5         8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "GRM4         8.643856    9.516300    8.655169   11.837607   11.610034  \n",
       "H19          8.643856    8.643856   10.807397    9.956303    8.643856  \n",
       "HAR1A        8.643856    8.643856    8.643856    9.955595   10.487835  \n",
       "HAR1B        8.643856    8.643856   10.168098    9.302220    9.765537  \n",
       "HOTAIR       8.643856    8.643856    8.643856    8.643856   10.070974  \n",
       "HOTAIRM1     8.643856    8.643856    8.949605    8.643856    8.643856  \n",
       "HOTTIP       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "HULC         8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "IPW          8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "JPX         10.462688   11.692138   11.618579    8.643856    8.643856  \n",
       "KRASP1       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "KCNQ1OT1     8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "MEG3         8.643856   10.799453   12.503942   13.704860   13.976352  \n",
       "MESTIT1     10.120307   11.617112   13.244554   13.167716   13.273113  \n",
       "MALAT1       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "NEAT1        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "NRON         8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "NTT          8.643856    8.643856   11.185743   10.335539    8.643856  \n",
       "PCGEM1       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "PRINS        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "PRNCR1       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "PTENP1       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "SNHG1        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "SNHG3        8.643856    8.643856    8.643856   10.601943    8.643856  \n",
       "SNHG4        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "SNHG5        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "SNHG6        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "SPRY4-IT1    8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "TUG1         8.643856    8.643856   11.150089   10.231673    8.643856  \n",
       "UBC          8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "UCA1         8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "WT1-AS       8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "XIST         8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "ZFAS1        8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "MASCRNA      8.643856    8.643856    8.643856    8.643856    8.643856  \n",
       "\n",
       "[44 rows x 175 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hepatitis</th>\n",
       "      <th>7SK</th>\n",
       "      <th>AIR</th>\n",
       "      <th>COPG2IT1</th>\n",
       "      <th>DGCR5</th>\n",
       "      <th>DISC2</th>\n",
       "      <th>DIO3OS</th>\n",
       "      <th>EMX2OS</th>\n",
       "      <th>GAPDH</th>\n",
       "      <th>GAS5</th>\n",
       "      <th>...</th>\n",
       "      <th>SNHG5</th>\n",
       "      <th>SNHG6</th>\n",
       "      <th>SPRY4-IT1</th>\n",
       "      <th>TUG1</th>\n",
       "      <th>UBC</th>\n",
       "      <th>UCA1</th>\n",
       "      <th>WT1-AS</th>\n",
       "      <th>XIST</th>\n",
       "      <th>ZFAS1</th>\n",
       "      <th>MASCRNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM2068378</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.603117</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.094534</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068379</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.847305</td>\n",
       "      <td>9.750400</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.929325</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.537742</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.070961</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068380</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.266793</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.686277</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068381</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.977783</td>\n",
       "      <td>10.825720</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.094759</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.584173</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.807389</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068382</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.383699</td>\n",
       "      <td>10.434356</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.915263</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.892180</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068548</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.477524</td>\n",
       "      <td>9.290551</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068549</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.104535</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>9.033588</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068550</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>12.809209</td>\n",
       "      <td>10.758394</td>\n",
       "      <td>8.959334</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.150089</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068551</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>11.095331</td>\n",
       "      <td>10.532171</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.304445</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.231673</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM2068552</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>10.007780</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "      <td>8.643856</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Hepatitis       7SK       AIR  COPG2IT1      DGCR5      DISC2  \\\n",
       "GSM2068378        0.0  8.643856  8.643856  9.603117   8.643856  11.094534   \n",
       "GSM2068379        0.0  8.643856  8.643856  8.643856  11.847305   9.750400   \n",
       "GSM2068380        0.0  8.643856  8.643856  8.643856  10.266793   8.643856   \n",
       "GSM2068381        0.0  8.643856  8.643856  8.643856  11.977783  10.825720   \n",
       "GSM2068382        0.0  8.643856  8.643856  8.643856  11.383699  10.434356   \n",
       "...               ...       ...       ...       ...        ...        ...   \n",
       "GSM2068548        0.0  8.643856  8.643856  8.643856   9.477524   9.290551   \n",
       "GSM2068549        0.0  8.643856  8.643856  8.643856  11.104535   8.643856   \n",
       "GSM2068550        0.0  8.643856  8.643856  8.643856  12.809209  10.758394   \n",
       "GSM2068551        0.0  8.643856  8.643856  8.643856  11.095331  10.532171   \n",
       "GSM2068552        0.0  8.643856  8.643856  8.643856   8.643856   8.643856   \n",
       "\n",
       "              DIO3OS    EMX2OS      GAPDH      GAS5  ...     SNHG5     SNHG6  \\\n",
       "GSM2068378  8.643856  8.643856   8.643856  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068379  8.643856  8.643856   8.929325  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068380  8.643856  8.643856   8.686277  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068381  8.643856  8.643856   9.094759  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068382  8.643856  8.643856   8.915263  8.643856  ...  8.643856  8.643856   \n",
       "...              ...       ...        ...       ...  ...       ...       ...   \n",
       "GSM2068548  8.643856  8.643856   8.643856  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068549  8.643856  8.643856   9.033588  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068550  8.959334  8.643856   8.643856  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068551  8.643856  8.643856  10.304445  8.643856  ...  8.643856  8.643856   \n",
       "GSM2068552  8.643856  8.643856  10.007780  8.643856  ...  8.643856  8.643856   \n",
       "\n",
       "            SPRY4-IT1       TUG1       UBC      UCA1    WT1-AS       XIST  \\\n",
       "GSM2068378   8.643856   8.643856  8.643856  8.643856  8.643856   8.643856   \n",
       "GSM2068379   8.643856  11.537742  8.643856  8.643856  8.643856  10.070961   \n",
       "GSM2068380   8.643856   8.643856  8.643856  8.643856  8.643856   8.643856   \n",
       "GSM2068381   8.643856  11.584173  8.643856  8.643856  8.643856   8.807389   \n",
       "GSM2068382   8.643856   8.892180  8.643856  8.643856  8.643856   8.643856   \n",
       "...               ...        ...       ...       ...       ...        ...   \n",
       "GSM2068548   8.643856   8.643856  8.643856  8.643856  8.643856   8.643856   \n",
       "GSM2068549   8.643856   8.643856  8.643856  8.643856  8.643856   8.643856   \n",
       "GSM2068550   8.643856  11.150089  8.643856  8.643856  8.643856   8.643856   \n",
       "GSM2068551   8.643856  10.231673  8.643856  8.643856  8.643856   8.643856   \n",
       "GSM2068552   8.643856   8.643856  8.643856  8.643856  8.643856   8.643856   \n",
       "\n",
       "               ZFAS1   MASCRNA  \n",
       "GSM2068378  8.643856  8.643856  \n",
       "GSM2068379  8.643856  8.643856  \n",
       "GSM2068380  8.643856  8.643856  \n",
       "GSM2068381  8.643856  8.643856  \n",
       "GSM2068382  8.643856  8.643856  \n",
       "...              ...       ...  \n",
       "GSM2068548  8.643856  8.643856  \n",
       "GSM2068549  8.643856  8.643856  \n",
       "GSM2068550  8.643856  8.643856  \n",
       "GSM2068551  8.643856  8.643856  \n",
       "GSM2068552  8.643856  8.643856  \n",
       "\n",
       "[175 rows x 45 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 175 samples.\n",
      "For the feature 'Hepatitis', the least common label is '1.0' with 45 occurrences. This represents 25.71% of the dataset.\n",
      "The distribution of the feature 'Hepatitis' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output\\Jiayi\\Hepatitis\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Molecular signature predictive of long-term liver fibrosis progression to inform anti-fibrotic drug development\"\n",
      "!Series_summary\t\"This SuperSeries is composed of the SubSeries listed below.\"\n",
      "!Series_overall_design\t\"Refer to individual Series\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM4557370</th>\n",
       "      <th>GSM4557371</th>\n",
       "      <th>GSM4557372</th>\n",
       "      <th>GSM4557373</th>\n",
       "      <th>GSM4557374</th>\n",
       "      <th>GSM4557375</th>\n",
       "      <th>GSM4557376</th>\n",
       "      <th>GSM4557377</th>\n",
       "      <th>GSM4557378</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5517662</th>\n",
       "      <th>GSM5517663</th>\n",
       "      <th>GSM5517664</th>\n",
       "      <th>GSM5517665</th>\n",
       "      <th>GSM5517666</th>\n",
       "      <th>GSM5517667</th>\n",
       "      <th>GSM5517668</th>\n",
       "      <th>GSM5517669</th>\n",
       "      <th>GSM5517670</th>\n",
       "      <th>GSM5517671</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>fibrosis stage: 0</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "      <td>tissue: Liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>pls risk prediction: High</td>\n",
       "      <td>pls risk prediction: Intermediate</td>\n",
       "      <td>pls risk prediction: Intermediate</td>\n",
       "      <td>pls risk prediction: High</td>\n",
       "      <td>pls risk prediction: Intermediate</td>\n",
       "      <td>pls risk prediction: High</td>\n",
       "      <td>pls risk prediction: High</td>\n",
       "      <td>pls risk prediction: Intermediate</td>\n",
       "      <td>pls risk prediction: High</td>\n",
       "      <td>...</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "      <td>sample group: Vehicle control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "      <td>compound: DMSO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "      <td>concentration: 0.1%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 653 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                 GSM4557370  \\\n",
       "0  !Sample_characteristics_ch1          fibrosis stage: 0   \n",
       "1  !Sample_characteristics_ch1  pls risk prediction: High   \n",
       "2  !Sample_characteristics_ch1                        NaN   \n",
       "3  !Sample_characteristics_ch1                        NaN   \n",
       "\n",
       "                          GSM4557371                         GSM4557372  \\\n",
       "0                  fibrosis stage: 0                  fibrosis stage: 0   \n",
       "1  pls risk prediction: Intermediate  pls risk prediction: Intermediate   \n",
       "2                                NaN                                NaN   \n",
       "3                                NaN                                NaN   \n",
       "\n",
       "                  GSM4557373                         GSM4557374  \\\n",
       "0          fibrosis stage: 0                  fibrosis stage: 0   \n",
       "1  pls risk prediction: High  pls risk prediction: Intermediate   \n",
       "2                        NaN                                NaN   \n",
       "3                        NaN                                NaN   \n",
       "\n",
       "                  GSM4557375                 GSM4557376  \\\n",
       "0          fibrosis stage: 0          fibrosis stage: 0   \n",
       "1  pls risk prediction: High  pls risk prediction: High   \n",
       "2                        NaN                        NaN   \n",
       "3                        NaN                        NaN   \n",
       "\n",
       "                          GSM4557377                 GSM4557378  ...  \\\n",
       "0                  fibrosis stage: 0          fibrosis stage: 0  ...   \n",
       "1  pls risk prediction: Intermediate  pls risk prediction: High  ...   \n",
       "2                                NaN                        NaN  ...   \n",
       "3                                NaN                        NaN  ...   \n",
       "\n",
       "                      GSM5517662                     GSM5517663  \\\n",
       "0                  tissue: Liver                  tissue: Liver   \n",
       "1  sample group: Vehicle control  sample group: Vehicle control   \n",
       "2                 compound: DMSO                 compound: DMSO   \n",
       "3            concentration: 0.1%            concentration: 0.1%   \n",
       "\n",
       "                      GSM5517664                     GSM5517665  \\\n",
       "0                  tissue: Liver                  tissue: Liver   \n",
       "1  sample group: Vehicle control  sample group: Vehicle control   \n",
       "2                 compound: DMSO                 compound: DMSO   \n",
       "3            concentration: 0.1%            concentration: 0.1%   \n",
       "\n",
       "                      GSM5517666                     GSM5517667  \\\n",
       "0                  tissue: Liver                  tissue: Liver   \n",
       "1  sample group: Vehicle control  sample group: Vehicle control   \n",
       "2                 compound: DMSO                 compound: DMSO   \n",
       "3            concentration: 0.1%            concentration: 0.1%   \n",
       "\n",
       "                      GSM5517668                     GSM5517669  \\\n",
       "0                  tissue: Liver                  tissue: Liver   \n",
       "1  sample group: Vehicle control  sample group: Vehicle control   \n",
       "2                 compound: DMSO                 compound: DMSO   \n",
       "3            concentration: 0.1%            concentration: 0.1%   \n",
       "\n",
       "                      GSM5517670                     GSM5517671  \n",
       "0                  tissue: Liver                  tissue: Liver  \n",
       "1  sample group: Vehicle control  sample group: Vehicle control  \n",
       "2                 compound: DMSO                 compound: DMSO  \n",
       "3            concentration: 0.1%            concentration: 0.1%  \n",
       "\n",
       "[4 rows x 653 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No obvious trait convert\n",
    "cohort = accession_num = \"GSE85550\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/zhangmianchen/Downloads/DATA/GEO/Hepatitis/GSE113392/GSE113392_family.soft.gz',\n",
       " '/Users/zhangmianchen/Downloads/DATA/GEO/Hepatitis/GSE113392/GSE113392_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = accession_num = \"GSE113392\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Loci-specific differences in blood DNA methylation in HBV-negative populations at risk for hepatocellular carcinoma development - pre-diagnostic HCC blood samples\"\n",
      "!Series_summary\t\"Genome wide DNA methylation profiling of blood samples collected from patients prior to diagnosis with hepatocellular carcinoma (HCC) vs. blood samples collected from healthy individuals without family history of cancer. The Illumina Infinium 450K Human DNA methylation Beadchip v1.2 was used to obtain DNA methylation profiles across approximately 450,000 CpGs in human samples corresponding to cases (pre-diagnostic HCC) and controls. Samples included 21 cases and 21 controls. Cases were matched with controls on gender, age, ethnicity, hepatitis C infection, and diabetes. The presence or absence of HCC in our study was determined based on the AASLD criteria.\"\n",
      "!Series_overall_design\t\"Bisulphite converted DNA from the 42 samples were hybridised to the Illumina Infinium 450k Human Methylation Beadchip v1.2\"\n"
     ]
    }
   ],
   "source": [
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM3104805</th>\n",
       "      <th>GSM3104806</th>\n",
       "      <th>GSM3104807</th>\n",
       "      <th>GSM3104808</th>\n",
       "      <th>GSM3104809</th>\n",
       "      <th>GSM3104810</th>\n",
       "      <th>GSM3104811</th>\n",
       "      <th>GSM3104812</th>\n",
       "      <th>GSM3104813</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3104837</th>\n",
       "      <th>GSM3104838</th>\n",
       "      <th>GSM3104839</th>\n",
       "      <th>GSM3104840</th>\n",
       "      <th>GSM3104841</th>\n",
       "      <th>GSM3104842</th>\n",
       "      <th>GSM3104843</th>\n",
       "      <th>GSM3104844</th>\n",
       "      <th>GSM3104845</th>\n",
       "      <th>GSM3104846</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "      <td>case/control: Control - Healthy</td>\n",
       "      <td>case/control: Case - pre-diagnostic hepatocell...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>...</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "      <td>sample type: human whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "      <td>disease state: normal</td>\n",
       "      <td>disease state: hepatocellular carcinoma (HCC)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                       GSM3104805  \\\n",
       "0  !Sample_characteristics_ch1  case/control: Control - Healthy   \n",
       "1  !Sample_characteristics_ch1   sample type: human whole blood   \n",
       "2  !Sample_characteristics_ch1            disease state: normal   \n",
       "\n",
       "                                          GSM3104806  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104807  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104808  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104809  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104810  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104811  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104812  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104813  ...                       GSM3104837  \\\n",
       "0  case/control: Control - Healthy  ...  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood  ...   sample type: human whole blood   \n",
       "2            disease state: normal  ...            disease state: normal   \n",
       "\n",
       "                                          GSM3104838  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104839  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104840  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104841  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104842  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104843  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104844  \\\n",
       "0  case/control: Case - pre-diagnostic hepatocell...   \n",
       "1                     sample type: human whole blood   \n",
       "2      disease state: hepatocellular carcinoma (HCC)   \n",
       "\n",
       "                        GSM3104845  \\\n",
       "0  case/control: Control - Healthy   \n",
       "1   sample type: human whole blood   \n",
       "2            disease state: normal   \n",
       "\n",
       "                                          GSM3104846  \n",
       "0  case/control: Case - pre-diagnostic hepatocell...  \n",
       "1                     sample type: human whole blood  \n",
       "2      disease state: hepatocellular carcinoma (HCC)  \n",
       "\n",
       "[3 rows x 43 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['case/control: Control - Healthy',\n",
       "  'case/control: Case - pre-diagnostic hepatocellular carcinoma (HCC)'],\n",
       " 1: ['sample type: human whole blood'],\n",
       " 2: ['disease state: normal', 'disease state: hepatocellular carcinoma (HCC)']}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Hepatitis\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Hepatitis\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Loci-specific differences in blood DNA methylation in HBV-negative populations at risk for hepatocellular carcinoma development - pre-diagnostic HCC blood samples\"\\n!Series_summary\\t\"Genome wide DNA methylation profiling of blood samples collected from patients prior to diagnosis with hepatocellular carcinoma (HCC) vs. blood samples collected from healthy individuals without family history of cancer. The Illumina Infinium 450K Human DNA methylation Beadchip v1.2 was used to obtain DNA methylation profiles across approximately 450,000 CpGs in human samples corresponding to cases (pre-diagnostic HCC) and controls. Samples included 21 cases and 21 controls. Cases were matched with controls on gender, age, ethnicity, hepatitis C infection, and diabetes. The presence or absence of HCC in our study was determined based on the AASLD criteria.\"\\n!Series_overall_design\\t\"Bisulphite converted DNA from the 42 samples were hybridised to the Illumina Infinium 450k Human Methylation Beadchip v1.2\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'case/control: Control - Healthy\\', \\'case/control: Case - pre-diagnostic hepatocellular carcinoma (HCC)\\'], 1: [\\'sample type: human whole blood\\'], 2: [\\'disease state: normal\\', \\'disease state: hepatocellular carcinoma (HCC)\\']}\\n'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'case/control: Case - pre-diagnostic hepatocellular carcinoma (HCC)':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if gender_string.lower() == 'male: 0':\n",
    "        return 0\n",
    "    elif gender_string.lower() == 'male: 1':\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3104805</th>\n",
       "      <th>GSM3104806</th>\n",
       "      <th>GSM3104807</th>\n",
       "      <th>GSM3104808</th>\n",
       "      <th>GSM3104809</th>\n",
       "      <th>GSM3104810</th>\n",
       "      <th>GSM3104811</th>\n",
       "      <th>GSM3104812</th>\n",
       "      <th>GSM3104813</th>\n",
       "      <th>GSM3104814</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3104837</th>\n",
       "      <th>GSM3104838</th>\n",
       "      <th>GSM3104839</th>\n",
       "      <th>GSM3104840</th>\n",
       "      <th>GSM3104841</th>\n",
       "      <th>GSM3104842</th>\n",
       "      <th>GSM3104843</th>\n",
       "      <th>GSM3104844</th>\n",
       "      <th>GSM3104845</th>\n",
       "      <th>GSM3104846</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hepatitis</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM3104805  GSM3104806  GSM3104807  GSM3104808  GSM3104809  \\\n",
       "Hepatitis           0           1           0           1           0   \n",
       "\n",
       "           GSM3104810  GSM3104811  GSM3104812  GSM3104813  GSM3104814  ...  \\\n",
       "Hepatitis           1           0           1           0           1  ...   \n",
       "\n",
       "           GSM3104837  GSM3104838  GSM3104839  GSM3104840  GSM3104841  \\\n",
       "Hepatitis           0           1           0           1           0   \n",
       "\n",
       "           GSM3104842  GSM3104843  GSM3104844  GSM3104845  GSM3104846  \n",
       "Hepatitis           1           0           1           0           1  \n",
       "\n",
       "[1 rows x 42 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3104805</th>\n",
       "      <th>GSM3104806</th>\n",
       "      <th>GSM3104807</th>\n",
       "      <th>GSM3104808</th>\n",
       "      <th>GSM3104809</th>\n",
       "      <th>GSM3104810</th>\n",
       "      <th>GSM3104811</th>\n",
       "      <th>GSM3104812</th>\n",
       "      <th>GSM3104813</th>\n",
       "      <th>GSM3104814</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM3104837</th>\n",
       "      <th>GSM3104838</th>\n",
       "      <th>GSM3104839</th>\n",
       "      <th>GSM3104840</th>\n",
       "      <th>GSM3104841</th>\n",
       "      <th>GSM3104842</th>\n",
       "      <th>GSM3104843</th>\n",
       "      <th>GSM3104844</th>\n",
       "      <th>GSM3104845</th>\n",
       "      <th>GSM3104846</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000029</th>\n",
       "      <td>0.641753</td>\n",
       "      <td>0.536023</td>\n",
       "      <td>0.480923</td>\n",
       "      <td>0.413408</td>\n",
       "      <td>0.503781</td>\n",
       "      <td>0.484459</td>\n",
       "      <td>0.322564</td>\n",
       "      <td>0.598879</td>\n",
       "      <td>0.472014</td>\n",
       "      <td>0.492969</td>\n",
       "      <td>...</td>\n",
       "      <td>0.469127</td>\n",
       "      <td>0.473445</td>\n",
       "      <td>0.488830</td>\n",
       "      <td>0.434073</td>\n",
       "      <td>0.543320</td>\n",
       "      <td>0.533835</td>\n",
       "      <td>0.532300</td>\n",
       "      <td>0.554394</td>\n",
       "      <td>0.654268</td>\n",
       "      <td>0.534820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000108</th>\n",
       "      <td>0.946797</td>\n",
       "      <td>0.964972</td>\n",
       "      <td>0.949461</td>\n",
       "      <td>0.941012</td>\n",
       "      <td>0.917665</td>\n",
       "      <td>0.917262</td>\n",
       "      <td>0.945542</td>\n",
       "      <td>0.949976</td>\n",
       "      <td>0.922315</td>\n",
       "      <td>0.953664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.953520</td>\n",
       "      <td>0.942573</td>\n",
       "      <td>0.935418</td>\n",
       "      <td>0.947862</td>\n",
       "      <td>0.955645</td>\n",
       "      <td>0.942716</td>\n",
       "      <td>0.942501</td>\n",
       "      <td>0.948831</td>\n",
       "      <td>0.954096</td>\n",
       "      <td>0.912201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000109</th>\n",
       "      <td>0.898792</td>\n",
       "      <td>0.904099</td>\n",
       "      <td>0.908054</td>\n",
       "      <td>0.890208</td>\n",
       "      <td>0.883757</td>\n",
       "      <td>0.861751</td>\n",
       "      <td>0.897760</td>\n",
       "      <td>0.881591</td>\n",
       "      <td>0.899898</td>\n",
       "      <td>0.913731</td>\n",
       "      <td>...</td>\n",
       "      <td>0.915518</td>\n",
       "      <td>0.880797</td>\n",
       "      <td>0.911363</td>\n",
       "      <td>0.786655</td>\n",
       "      <td>0.865565</td>\n",
       "      <td>0.890660</td>\n",
       "      <td>0.898855</td>\n",
       "      <td>0.890524</td>\n",
       "      <td>0.836090</td>\n",
       "      <td>0.878666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000165</th>\n",
       "      <td>0.244225</td>\n",
       "      <td>0.203119</td>\n",
       "      <td>0.276537</td>\n",
       "      <td>0.140747</td>\n",
       "      <td>0.124542</td>\n",
       "      <td>0.132241</td>\n",
       "      <td>0.149900</td>\n",
       "      <td>0.146812</td>\n",
       "      <td>0.208076</td>\n",
       "      <td>0.171069</td>\n",
       "      <td>...</td>\n",
       "      <td>0.232690</td>\n",
       "      <td>0.210047</td>\n",
       "      <td>0.209973</td>\n",
       "      <td>0.155052</td>\n",
       "      <td>0.168326</td>\n",
       "      <td>0.162929</td>\n",
       "      <td>0.156828</td>\n",
       "      <td>0.211841</td>\n",
       "      <td>0.196233</td>\n",
       "      <td>0.136576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000236</th>\n",
       "      <td>0.776429</td>\n",
       "      <td>0.809852</td>\n",
       "      <td>0.809300</td>\n",
       "      <td>0.767984</td>\n",
       "      <td>0.737150</td>\n",
       "      <td>0.728721</td>\n",
       "      <td>0.732519</td>\n",
       "      <td>0.869120</td>\n",
       "      <td>0.806040</td>\n",
       "      <td>0.749944</td>\n",
       "      <td>...</td>\n",
       "      <td>0.730795</td>\n",
       "      <td>0.833923</td>\n",
       "      <td>0.771805</td>\n",
       "      <td>0.707987</td>\n",
       "      <td>0.824090</td>\n",
       "      <td>0.767424</td>\n",
       "      <td>0.734655</td>\n",
       "      <td>0.814077</td>\n",
       "      <td>0.786997</td>\n",
       "      <td>0.774174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM3104805  GSM3104806  GSM3104807  GSM3104808  GSM3104809  \\\n",
       "ID                                                                       \n",
       "cg00000029    0.641753    0.536023    0.480923    0.413408    0.503781   \n",
       "cg00000108    0.946797    0.964972    0.949461    0.941012    0.917665   \n",
       "cg00000109    0.898792    0.904099    0.908054    0.890208    0.883757   \n",
       "cg00000165    0.244225    0.203119    0.276537    0.140747    0.124542   \n",
       "cg00000236    0.776429    0.809852    0.809300    0.767984    0.737150   \n",
       "\n",
       "            GSM3104810  GSM3104811  GSM3104812  GSM3104813  GSM3104814  ...  \\\n",
       "ID                                                                      ...   \n",
       "cg00000029    0.484459    0.322564    0.598879    0.472014    0.492969  ...   \n",
       "cg00000108    0.917262    0.945542    0.949976    0.922315    0.953664  ...   \n",
       "cg00000109    0.861751    0.897760    0.881591    0.899898    0.913731  ...   \n",
       "cg00000165    0.132241    0.149900    0.146812    0.208076    0.171069  ...   \n",
       "cg00000236    0.728721    0.732519    0.869120    0.806040    0.749944  ...   \n",
       "\n",
       "            GSM3104837  GSM3104838  GSM3104839  GSM3104840  GSM3104841  \\\n",
       "ID                                                                       \n",
       "cg00000029    0.469127    0.473445    0.488830    0.434073    0.543320   \n",
       "cg00000108    0.953520    0.942573    0.935418    0.947862    0.955645   \n",
       "cg00000109    0.915518    0.880797    0.911363    0.786655    0.865565   \n",
       "cg00000165    0.232690    0.210047    0.209973    0.155052    0.168326   \n",
       "cg00000236    0.730795    0.833923    0.771805    0.707987    0.824090   \n",
       "\n",
       "            GSM3104842  GSM3104843  GSM3104844  GSM3104845  GSM3104846  \n",
       "ID                                                                      \n",
       "cg00000029    0.533835    0.532300    0.554394    0.654268    0.534820  \n",
       "cg00000108    0.942716    0.942501    0.948831    0.954096    0.912201  \n",
       "cg00000109    0.890660    0.898855    0.890524    0.836090    0.878666  \n",
       "cg00000165    0.162929    0.156828    0.211841    0.196233    0.136576  \n",
       "cg00000236    0.767424    0.734655    0.814077    0.786997    0.774174  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cg00000029',\n",
       " 'cg00000108',\n",
       " 'cg00000109',\n",
       " 'cg00000165',\n",
       " 'cg00000236',\n",
       " 'cg00000289',\n",
       " 'cg00000292',\n",
       " 'cg00000321',\n",
       " 'cg00000363',\n",
       " 'cg00000622',\n",
       " 'cg00000658',\n",
       " 'cg00000714',\n",
       " 'cg00000721',\n",
       " 'cg00000734',\n",
       " 'cg00000769',\n",
       " 'cg00000807',\n",
       " 'cg00000884',\n",
       " 'cg00000905',\n",
       " 'cg00000924',\n",
       " 'cg00000948']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['cg00000029', 'cg00000108', 'cg00000109', 'cg00000165', 'cg00000236', 'cg00000289', 'cg00000292', 'cg00000321', 'cg00000363', 'cg00000622', 'cg00000658', 'cg00000714', 'cg00000721', 'cg00000734', 'cg00000769', 'cg00000807', 'cg00000884', 'cg00000905', 'cg00000924', 'cg00000948']\\n\""
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'UCSC_RefGene_Name'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['cg00000029', 'cg00000108', 'cg00000109', 'cg00000165', 'cg00000236', 'cg00000289', 'cg00000292', '\n",
      "1000 input query terms found no hit:\t['cg00040862', 'cg00040873', 'cg00040923', 'cg00040945', 'cg00040986', 'cg00041014', 'cg00041047', '\n",
      "1000 input query terms found no hit:\t['cg00087971', 'cg00088007', 'cg00088026', 'cg00088033', 'cg00088042', 'cg00088079', 'cg00088080', '\n",
      "1000 input query terms found no hit:\t['cg00134190', 'cg00134210', 'cg00134212', 'cg00134295', 'cg00134340', 'cg00134415', 'cg00134539', '\n",
      "1000 input query terms found no hit:\t['cg00182202', 'cg00182273', 'cg00182288', 'cg00182294', 'cg00182416', 'cg00182421', 'cg00182534', '\n",
      "1000 input query terms found no hit:\t['cg00237631', 'cg00237752', 'cg00237825', 'cg00237876', 'cg00237904', 'cg00238038', 'cg00238052', '\n",
      "1000 input query terms found no hit:\t['cg00290039', 'cg00290086', 'cg00290101', 'cg00290158', 'cg00290218', 'cg00290303', 'cg00290340', '\n",
      "1000 input query terms found no hit:\t['cg00336605', 'cg00336669', 'cg00336700', 'cg00336706', 'cg00336791', 'cg00336800', 'cg00336867', '\n",
      "1000 input query terms found no hit:\t['cg00384539', 'cg00384553', 'cg00384577', 'cg00384632', 'cg00384653', 'cg00384701', 'cg00384707', '\n",
      "1000 input query terms found no hit:\t['cg00431835', 'cg00431837', 'cg00431894', 'cg00431924', 'cg00431968', 'cg00432041', 'cg00432049', '\n",
      "1000 input query terms found no hit:\t['cg00478692', 'cg00478710', 'cg00478851', 'cg00478874', 'cg00478884', 'cg00479101', 'cg00479208', '\n",
      "1000 input query terms found no hit:\t['cg00525503', 'cg00525508', 'cg00525597', 'cg00525681', 'cg00525683', 'cg00525722', 'cg00525823', '\n",
      "1000 input query terms found no hit:\t['cg00574014', 'cg00574017', 'cg00574114', 'cg00574196', 'cg00574206', 'cg00574213', 'cg00574251', '\n",
      "1000 input query terms found no hit:\t['cg00620921', 'cg00620976', 'cg00621029', 'cg00621084', 'cg00621125', 'cg00621197', 'cg00621240', '\n",
      "1000 input query terms found no hit:\t['cg00674003', 'cg00674004', 'cg00674024', 'cg00674043', 'cg00674220', 'cg00674265', 'cg00674365', '\n",
      "1000 input query terms found no hit:\t['cg00720822', 'cg00720829', 'cg00720839', 'cg00720845', 'cg00721016', 'cg00721080', 'cg00721086', '\n",
      "1000 input query terms found no hit:\t['cg00771026', 'cg00771049', 'cg00771084', 'cg00771108', 'cg00771170', 'cg00771185', 'cg00771217', '\n",
      "1000 input query terms found no hit:\t['cg00819233', 'cg00819310', 'cg00819338', 'cg00819362', 'cg00819619', 'cg00819696', 'cg00819788', '\n",
      "1000 input query terms found no hit:\t['cg00870166', 'cg00870179', 'cg00870242', 'cg00870265', 'cg00870269', 'cg00870279', 'cg00870449', '\n",
      "1000 input query terms found no hit:\t['cg00920327', 'cg00920337', 'cg00920348', 'cg00920395', 'cg00920422', 'cg00920424', 'cg00920477', '\n",
      "1000 input query terms found no hit:\t['cg00968616', 'cg00968638', 'cg00968640', 'cg00968771', 'cg00968800', 'cg00968875', 'cg00968890', '\n",
      "1000 input query terms found no hit:\t['cg01015769', 'cg01015871', 'cg01015879', 'cg01015899', 'cg01016030', 'cg01016090', 'cg01016092', '\n",
      "1000 input query terms found no hit:\t['cg01062068', 'cg01062113', 'cg01062116', 'cg01062126', 'cg01062247', 'cg01062269', 'cg01062395', '\n",
      "1000 input query terms found no hit:\t['cg01108614', 'cg01108633', 'cg01108642', 'cg01108697', 'cg01108705', 'cg01108710', 'cg01108713', '\n",
      "1000 input query terms found no hit:\t['cg01159004', 'cg01159097', 'cg01159149', 'cg01159194', 'cg01159288', 'cg01159294', 'cg01159380', '\n",
      "1000 input query terms found no hit:\t['cg01210264', 'cg01210288', 'cg01210307', 'cg01210372', 'cg01210520', 'cg01210550', 'cg01210554', '\n",
      "1000 input query terms found no hit:\t['cg01263541', 'cg01263574', 'cg01263575', 'cg01263624', 'cg01263631', 'cg01263691', 'cg01263716', '\n",
      "1000 input query terms found no hit:\t['cg01313176', 'cg01313234', 'cg01313313', 'cg01313320', 'cg01313358', 'cg01313363', 'cg01313387', '\n",
      "1000 input query terms found no hit:\t['cg01369479', 'cg01369511', 'cg01369522', 'cg01369646', 'cg01369689', 'cg01369750', 'cg01369819', '\n",
      "1000 input query terms found no hit:\t['cg01418256', 'cg01418261', 'cg01418291', 'cg01418351', 'cg01418385', 'cg01418388', 'cg01418441', '\n",
      "1000 input query terms found no hit:\t['cg01466348', 'cg01466491', 'cg01466576', 'cg01466615', 'cg01466631', 'cg01466678', 'cg01466741', '\n",
      "1000 input query terms found no hit:\t['cg01521648', 'cg01521674', 'cg01521688', 'cg01521987', 'cg01522083', 'cg01522095', 'cg01522243', '\n",
      "1000 input query terms found no hit:\t['cg01572940', 'cg01572979', 'cg01572989', 'cg01573067', 'cg01573089', 'cg01573121', 'cg01573140', '\n",
      "1000 input query terms found no hit:\t['cg01623663', 'cg01623718', 'cg01623771', 'cg01623797', 'cg01623806', 'cg01623825', 'cg01623879', '\n",
      "1000 input query terms found no hit:\t['cg01683969', 'cg01683986', 'cg01684006', 'cg01684175', 'cg01684233', 'cg01684248', 'cg01684256', '\n",
      "1000 input query terms found no hit:\t['cg01738457', 'cg01738522', 'cg01738524', 'cg01738593', 'cg01738638', 'cg01738650', 'cg01738805', '\n",
      "1000 input query terms found no hit:\t['cg01789926', 'cg01789962', 'cg01789978', 'cg01789980', 'cg01790002', 'cg01790032', 'cg01790083', '\n",
      "1000 input query terms found no hit:\t['cg01834225', 'cg01834305', 'cg01834373', 'cg01834528', 'cg01834541', 'cg01834566', 'cg01834573', '\n",
      "1000 input query terms found no hit:\t['cg01886741', 'cg01886810', 'cg01886855', 'cg01886893', 'cg01886957', 'cg01886988', 'cg01887138', '\n",
      "1000 input query terms found no hit:\t['cg01938066', 'cg01938118', 'cg01938295', 'cg01938304', 'cg01938354', 'cg01938359', 'cg01938422', '\n",
      "1000 input query terms found no hit:\t['cg01985201', 'cg01985343', 'cg01985390', 'cg01985396', 'cg01985412', 'cg01985446', 'cg01985465', '\n",
      "1000 input query terms found no hit:\t['cg02035565', 'cg02035603', 'cg02035605', 'cg02035681', 'cg02035698', 'cg02035733', 'cg02035751', '\n",
      "1000 input query terms found no hit:\t['cg02082887', 'cg02082929', 'cg02083029', 'cg02083080', 'cg02083109', 'cg02083151', 'cg02083189', '\n",
      "1000 input query terms found no hit:\t['cg02133941', 'cg02134022', 'cg02134046', 'cg02134189', 'cg02134194', 'cg02134200', 'cg02134248', '\n",
      "1000 input query terms found no hit:\t['cg02190038', 'cg02190127', 'cg02190211', 'cg02190253', 'cg02190350', 'cg02190353', 'cg02190383', '\n",
      "1000 input query terms found no hit:\t['cg02246229', 'cg02246426', 'cg02246427', 'cg02246609', 'cg02246640', 'cg02246645', 'cg02246665', '\n",
      "1000 input query terms found no hit:\t['cg02295389', 'cg02295453', 'cg02295504', 'cg02295509', 'cg02295525', 'cg02295527', 'cg02295574', '\n",
      "1000 input query terms found no hit:\t['cg02343451', 'cg02343467', 'cg02343503', 'cg02343602', 'cg02343604', 'cg02343628', 'cg02343648', '\n",
      "1000 input query terms found no hit:\t['cg02392176', 'cg02392179', 'cg02392228', 'cg02392248', 'cg02392269', 'cg02392395', 'cg02392508', '\n",
      "1000 input query terms found no hit:\t['cg02451122', 'cg02451362', 'cg02451443', 'cg02451467', 'cg02451502', 'cg02451508', 'cg02451516', '\n",
      "1000 input query terms found no hit:\t['cg02498311', 'cg02498382', 'cg02498422', 'cg02498484', 'cg02498579', 'cg02498702', 'cg02498722', '\n",
      "1000 input query terms found no hit:\t['cg02550602', 'cg02550607', 'cg02550691', 'cg02550722', 'cg02550738', 'cg02550741', 'cg02550896', '\n",
      "1000 input query terms found no hit:\t['cg02600347', 'cg02600349', 'cg02600394', 'cg02600411', 'cg02600413', 'cg02600430', 'cg02600478', '\n",
      "1000 input query terms found no hit:\t['cg02645977', 'cg02645993', 'cg02646021', 'cg02646064', 'cg02646073', 'cg02646225', 'cg02646251', '\n",
      "1000 input query terms found no hit:\t['cg02700479', 'cg02700491', 'cg02700498', 'cg02700606', 'cg02700624', 'cg02700714', 'cg02700824', '\n",
      "1000 input query terms found no hit:\t['cg02749073', 'cg02749088', 'cg02749105', 'cg02749148', 'cg02749215', 'cg02749356', 'cg02749463', '\n",
      "1000 input query terms found no hit:\t['cg02801598', 'cg02801645', 'cg02801772', 'cg02801786', 'cg02801789', 'cg02801839', 'cg02801993', '\n",
      "1000 input query terms found no hit:\t['cg02855778', 'cg02855833', 'cg02855836', 'cg02855850', 'cg02855883', 'cg02855981', 'cg02855983', '\n",
      "1000 input query terms found no hit:\t['cg02904062', 'cg02904064', 'cg02904093', 'cg02904184', 'cg02904235', 'cg02904285', 'cg02904344', '\n",
      "1000 input query terms found no hit:\t['cg02951717', 'cg02951719', 'cg02951738', 'cg02951739', 'cg02951765', 'cg02951766', 'cg02951847', '\n",
      "1000 input query terms found no hit:\t['cg02998028', 'cg02998075', 'cg02998129', 'cg02998155', 'cg02998206', 'cg02998240', 'cg02998278', '\n",
      "1000 input query terms found no hit:\t['cg03049846', 'cg03049863', 'cg03049881', 'cg03049917', 'cg03050022', 'cg03050096', 'cg03050127', '\n",
      "1000 input query terms found no hit:\t['cg03105070', 'cg03105091', 'cg03105133', 'cg03105159', 'cg03105222', 'cg03105244', 'cg03105348', '\n",
      "1000 input query terms found no hit:\t['cg03158194', 'cg03158218', 'cg03158310', 'cg03158314', 'cg03158400', 'cg03158499', 'cg03158561', '\n",
      "1000 input query terms found no hit:\t['cg03212862', 'cg03212952', 'cg03213202', 'cg03213216', 'cg03213259', 'cg03213289', 'cg03213352', '\n",
      "1000 input query terms found no hit:\t['cg03266527', 'cg03266576', 'cg03266686', 'cg03266780', 'cg03266904', 'cg03266942', 'cg03266952', '\n",
      "1000 input query terms found no hit:\t['cg03320170', 'cg03320255', 'cg03320372', 'cg03320441', 'cg03320492', 'cg03320521', 'cg03320607', '\n",
      "1000 input query terms found no hit:\t['cg03369779', 'cg03369807', 'cg03369828', 'cg03369906', 'cg03369957', 'cg03369965', 'cg03369999', '\n",
      "1000 input query terms found no hit:\t['cg03425468', 'cg03425504', 'cg03425506', 'cg03425609', 'cg03425615', 'cg03425695', 'cg03425731', '\n",
      "1000 input query terms found no hit:\t['cg03475182', 'cg03475245', 'cg03475285', 'cg03475293', 'cg03475307', 'cg03475420', 'cg03475491', '\n",
      "1000 input query terms found no hit:\t['cg03528710', 'cg03528784', 'cg03528824', 'cg03528848', 'cg03528946', 'cg03529003', 'cg03529015', '\n",
      "1000 input query terms found no hit:\t['cg03576419', 'cg03576422', 'cg03576425', 'cg03576467', 'cg03576469', 'cg03576480', 'cg03576530', '\n",
      "1000 input query terms found no hit:\t['cg03627015', 'cg03627022', 'cg03627261', 'cg03627290', 'cg03627295', 'cg03627300', 'cg03627309', '\n",
      "1000 input query terms found no hit:\t['cg03679398', 'cg03679456', 'cg03679504', 'cg03679521', 'cg03679544', 'cg03679565', 'cg03679581', '\n",
      "1000 input query terms found no hit:\t['cg03731131', 'cg03731171', 'cg03731202', 'cg03731235', 'cg03731251', 'cg03731268', 'cg03731303', '\n",
      "1000 input query terms found no hit:\t['cg03782584', 'cg03782662', 'cg03782727', 'cg03782771', 'cg03782778', 'cg03782800', 'cg03782845', '\n",
      "1000 input query terms found no hit:\t['cg03841686', 'cg03841705', 'cg03841784', 'cg03841832', 'cg03841837', 'cg03841977', 'cg03842063', '\n",
      "1000 input query terms found no hit:\t['cg03896661', 'cg03896685', 'cg03896694', 'cg03896752', 'cg03896792', 'cg03896836', 'cg03896940', '\n",
      "1000 input query terms found no hit:\t['cg03952578', 'cg03952724', 'cg03952885', 'cg03952984', 'cg03953069', 'cg03953077', 'cg03953086', '\n",
      "1000 input query terms found no hit:\t['cg04002486', 'cg04002608', 'cg04002618', 'cg04002758', 'cg04002794', 'cg04002822', 'cg04002885', '\n",
      "1000 input query terms found no hit:\t['cg04059318', 'cg04059367', 'cg04059454', 'cg04059461', 'cg04059568', 'cg04059647', 'cg04059696', '\n",
      "1000 input query terms found no hit:\t['cg04110544', 'cg04110559', 'cg04110601', 'cg04110665', 'cg04110708', 'cg04110750', 'cg04110868', '\n",
      "1000 input query terms found no hit:\t['cg04158367', 'cg04158589', 'cg04158612', 'cg04158698', 'cg04158730', 'cg04158792', 'cg04158862', '\n",
      "1000 input query terms found no hit:\t['cg04212340', 'cg04212450', 'cg04212500', 'cg04212568', 'cg04212619', 'cg04212651', 'cg04212699', '\n",
      "1000 input query terms found no hit:\t['cg04261408', 'cg04261496', 'cg04261529', 'cg04261556', 'cg04261607', 'cg04261665', 'cg04261791', '\n",
      "1000 input query terms found no hit:\t['cg04313565', 'cg04313601', 'cg04313639', 'cg04313733', 'cg04313735', 'cg04313756', 'cg04313758', '\n",
      "1000 input query terms found no hit:\t['cg04366104', 'cg04366249', 'cg04366343', 'cg04366367', 'cg04366381', 'cg04366385', 'cg04366407', '\n",
      "1000 input query terms found no hit:\t['cg04418355', 'cg04418423', 'cg04418434', 'cg04418492', 'cg04418529', 'cg04418576', 'cg04418585', '\n",
      "1000 input query terms found no hit:\t['cg04475027', 'cg04475037', 'cg04475069', 'cg04475085', 'cg04475095', 'cg04475157', 'cg04475193', '\n",
      "1000 input query terms found no hit:\t['cg04527876', 'cg04527918', 'cg04527989', 'cg04528024', 'cg04528038', 'cg04528043', 'cg04528054', '\n",
      "1000 input query terms found no hit:\t['cg04579183', 'cg04579203', 'cg04579211', 'cg04579254', 'cg04579263', 'cg04579398', 'cg04579415', '\n",
      "1000 input query terms found no hit:\t['cg04636920', 'cg04636967', 'cg04637007', 'cg04637129', 'cg04637204', 'cg04637264', 'cg04637372', '\n",
      "1000 input query terms found no hit:\t['cg04692960', 'cg04692993', 'cg04693014', 'cg04693046', 'cg04693256', 'cg04693329', 'cg04693379', '\n",
      "1000 input query terms found no hit:\t['cg04747996', 'cg04748010', 'cg04748051', 'cg04748098', 'cg04748223', 'cg04748321', 'cg04748357', '\n",
      "1000 input query terms found no hit:\t['cg04800768', 'cg04800788', 'cg04800816', 'cg04801003', 'cg04801085', 'cg04801098', 'cg04801133', '\n",
      "1000 input query terms found no hit:\t['cg04853925', 'cg04853958', 'cg04853978', 'cg04853995', 'cg04854025', 'cg04854089', 'cg04854098', '\n",
      "1000 input query terms found no hit:\t['cg04903089', 'cg04903116', 'cg04903146', 'cg04903157', 'cg04903159', 'cg04903163', 'cg04903360', '\n",
      "1000 input query terms found no hit:\t['cg04955051', 'cg04955097', 'cg04955210', 'cg04955231', 'cg04955246', 'cg04955303', 'cg04955333', '\n",
      "1000 input query terms found no hit:\t['cg05008218', 'cg05008247', 'cg05008296', 'cg05008344', 'cg05008366', 'cg05008367', 'cg05008386', '\n",
      "1000 input query terms found no hit:\t['cg05059994', 'cg05060175', 'cg05060427', 'cg05060440', 'cg05060486', 'cg05060534', 'cg05060577', '\n",
      "1000 input query terms found no hit:\t['cg05120028', 'cg05120042', 'cg05120074', 'cg05120085', 'cg05120113', 'cg05120150', 'cg05120156', '\n",
      "1000 input query terms found no hit:\t['cg05170353', 'cg05170366', 'cg05170390', 'cg05170445', 'cg05170452', 'cg05170483', 'cg05170494', '\n",
      "1000 input query terms found no hit:\t['cg05226168', 'cg05226180', 'cg05226242', 'cg05226283', 'cg05226292', 'cg05226322', 'cg05226335', '\n",
      "1000 input query terms found no hit:\t['cg05277096', 'cg05277122', 'cg05277141', 'cg05277146', 'cg05277165', 'cg05277329', 'cg05277407', '\n",
      "1000 input query terms found no hit:\t['cg05332152', 'cg05332270', 'cg05332306', 'cg05332308', 'cg05332483', 'cg05332489', 'cg05332525', '\n",
      "1000 input query terms found no hit:\t['cg05378190', 'cg05378199', 'cg05378295', 'cg05378452', 'cg05378609', 'cg05378690', 'cg05378695', '\n",
      "1000 input query terms found no hit:\t['cg05435882', 'cg05435993', 'cg05436007', 'cg05436043', 'cg05436064', 'cg05436155', 'cg05436231', '\n",
      "1000 input query terms found no hit:\t['cg05496768', 'cg05496814', 'cg05496988', 'cg05497107', 'cg05497175', 'cg05497199', 'cg05497216', '\n",
      "1000 input query terms found no hit:\t['cg05554396', 'cg05554457', 'cg05554494', 'cg05554543', 'cg05554592', 'cg05554594', 'cg05554718', '\n",
      "1000 input query terms found no hit:\t['cg05610632', 'cg05610699', 'cg05610740', 'cg05611007', 'cg05611177', 'cg05611199', 'cg05611218', '\n",
      "1000 input query terms found no hit:\t['cg05664352', 'cg05664421', 'cg05664472', 'cg05664523', 'cg05664581', 'cg05664613', 'cg05664664', '\n",
      "1000 input query terms found no hit:\t['cg05723825', 'cg05723953', 'cg05724065', 'cg05724097', 'cg05724110', 'cg05724197', 'cg05724240', '\n",
      "1000 input query terms found no hit:\t['cg05784148', 'cg05784157', 'cg05784171', 'cg05784193', 'cg05784201', 'cg05784226', 'cg05784314', '\n",
      "1000 input query terms found no hit:\t['cg05842815', 'cg05842855', 'cg05842890', 'cg05842943', 'cg05843106', 'cg05843257', 'cg05843312', '\n",
      "1000 input query terms found no hit:\t['cg05894907', 'cg05894910', 'cg05894948', 'cg05894970', 'cg05895018', 'cg05895034', 'cg05895077', '\n",
      "1000 input query terms found no hit:\t['cg05944004', 'cg05944137', 'cg05944173', 'cg05944174', 'cg05944193', 'cg05944207', 'cg05944260', '\n",
      "1000 input query terms found no hit:\t['cg05994320', 'cg05994458', 'cg05994508', 'cg05994672', 'cg05994690', 'cg05994794', 'cg05994819', '\n",
      "1000 input query terms found no hit:\t['cg06053605', 'cg06053651', 'cg06053702', 'cg06053738', 'cg06053805', 'cg06053850', 'cg06053862', '\n",
      "1000 input query terms found no hit:\t['cg06109284', 'cg06109379', 'cg06109482', 'cg06109529', 'cg06109624', 'cg06109664', 'cg06109724', '\n",
      "1000 input query terms found no hit:\t['cg06173216', 'cg06173319', 'cg06173395', 'cg06173416', 'cg06173461', 'cg06173470', 'cg06173481', '\n",
      "1000 input query terms found no hit:\t['cg06226630', 'cg06226670', 'cg06226675', 'cg06226703', 'cg06226724', 'cg06226973', 'cg06227019', '\n",
      "1000 input query terms found no hit:\t['cg06293099', 'cg06293108', 'cg06293169', 'cg06293170', 'cg06293172', 'cg06293180', 'cg06293195', '\n",
      "1000 input query terms found no hit:\t['cg06355830', 'cg06355870', 'cg06355908', 'cg06356032', 'cg06356308', 'cg06356333', 'cg06356454', '\n",
      "1000 input query terms found no hit:\t['cg06412323', 'cg06412358', 'cg06412409', 'cg06412434', 'cg06412435', 'cg06412465', 'cg06412523', '\n",
      "1000 input query terms found no hit:\t['cg06473261', 'cg06473276', 'cg06473288', 'cg06473363', 'cg06473429', 'cg06473449', 'cg06473539', '\n",
      "1000 input query terms found no hit:\t['cg06528601', 'cg06528626', 'cg06528645', 'cg06528737', 'cg06528773', 'cg06528803', 'cg06528816', '\n",
      "1000 input query terms found no hit:\t['cg06582266', 'cg06582394', 'cg06582411', 'cg06582512', 'cg06582575', 'cg06582663', 'cg06582706', '\n",
      "1000 input query terms found no hit:\t['cg06635722', 'cg06635764', 'cg06635797', 'cg06635832', 'cg06635849', 'cg06635871', 'cg06635900', '\n",
      "1000 input query terms found no hit:\t['cg06688293', 'cg06688323', 'cg06688362', 'cg06688396', 'cg06688399', 'cg06688409', 'cg06688411', '\n",
      "1000 input query terms found no hit:\t['cg06741568', 'cg06741594', 'cg06741653', 'cg06741726', 'cg06741803', 'cg06741896', 'cg06741909', '\n",
      "1000 input query terms found no hit:\t['cg06790053', 'cg06790069', 'cg06790171', 'cg06790197', 'cg06790275', 'cg06790305', 'cg06790324', '\n",
      "1000 input query terms found no hit:\t['cg06835777', 'cg06835786', 'cg06835814', 'cg06835822', 'cg06835960', 'cg06836020', 'cg06836032', '\n",
      "1000 input query terms found no hit:\t['cg06883949', 'cg06883956', 'cg06884029', 'cg06884199', 'cg06884254', 'cg06884262', 'cg06884304', '\n",
      "1000 input query terms found no hit:\t['cg06935666', 'cg06935856', 'cg06935967', 'cg06935979', 'cg06936079', 'cg06936108', 'cg06936155', '\n",
      "1000 input query terms found no hit:\t['cg06987369', 'cg06987468', 'cg06987504', 'cg06987616', 'cg06987623', 'cg06987672', 'cg06987881', '\n",
      "1000 input query terms found no hit:\t['cg07038001', 'cg07038098', 'cg07038187', 'cg07038191', 'cg07038243', 'cg07038342', 'cg07038400', '\n",
      "1000 input query terms found no hit:\t['cg07091102', 'cg07091154', 'cg07091207', 'cg07091220', 'cg07091346', 'cg07091355', 'cg07091392', '\n",
      "1000 input query terms found no hit:\t['cg07139317', 'cg07139329', 'cg07139330', 'cg07139350', 'cg07139363', 'cg07139438', 'cg07139440', '\n",
      "1000 input query terms found no hit:\t['cg07185038', 'cg07185041', 'cg07185068', 'cg07185119', 'cg07185131', 'cg07185165', 'cg07185174', '\n",
      "1000 input query terms found no hit:\t['cg07236874', 'cg07236884', 'cg07236904', 'cg07236911', 'cg07236912', 'cg07236925', 'cg07236939', '\n",
      "1000 input query terms found no hit:\t['cg07297111', 'cg07297178', 'cg07297242', 'cg07297322', 'cg07297397', 'cg07297410', 'cg07297467', '\n",
      "1000 input query terms found no hit:\t['cg07352798', 'cg07352834', 'cg07352859', 'cg07352917', 'cg07352967', 'cg07353006', 'cg07353253', '\n",
      "1000 input query terms found no hit:\t['cg07411420', 'cg07411432', 'cg07411620', 'cg07411754', 'cg07411798', 'cg07412043', 'cg07412079', '\n",
      "1000 input query terms found no hit:\t['cg07468956', 'cg07469063', 'cg07469105', 'cg07469215', 'cg07469402', 'cg07469408', 'cg07469445', '\n",
      "1000 input query terms found no hit:\t['cg07521394', 'cg07521483', 'cg07521506', 'cg07521618', 'cg07521668', 'cg07521723', 'cg07521789', '\n",
      "1000 input query terms found no hit:\t['cg07573366', 'cg07573461', 'cg07573715', 'cg07573717', 'cg07573727', 'cg07573786', 'cg07573797', '\n",
      "1000 input query terms found no hit:\t['cg07625156', 'cg07625177', 'cg07625194', 'cg07625271', 'cg07625383', 'cg07625430', 'cg07625546', '\n",
      "1000 input query terms found no hit:\t['cg07685977', 'cg07686035', 'cg07686122', 'cg07686248', 'cg07686392', 'cg07686394', 'cg07686441', '\n",
      "1000 input query terms found no hit:\t['cg07739188', 'cg07739205', 'cg07739215', 'cg07739257', 'cg07739369', 'cg07739398', 'cg07739472', '\n",
      "1000 input query terms found no hit:\t['cg07797907', 'cg07798131', 'cg07798137', 'cg07798143', 'cg07798295', 'cg07798347', 'cg07798380', '\n",
      "1000 input query terms found no hit:\t['cg07851269', 'cg07851278', 'cg07851290', 'cg07851361', 'cg07851374', 'cg07851521', 'cg07851632', '\n",
      "1000 input query terms found no hit:\t['cg07909749', 'cg07909759', 'cg07909897', 'cg07910050', 'cg07910065', 'cg07910408', 'cg07910460', '\n",
      "1000 input query terms found no hit:\t['cg07963725', 'cg07963741', 'cg07963760', 'cg07963816', 'cg07963854', 'cg07963988', 'cg07963990', '\n",
      "1000 input query terms found no hit:\t['cg08023458', 'cg08023476', 'cg08023527', 'cg08023692', 'cg08023751', 'cg08023796', 'cg08023809', '\n",
      "1000 input query terms found no hit:\t['cg08080060', 'cg08080145', 'cg08080174', 'cg08080195', 'cg08080213', 'cg08080334', 'cg08080395', '\n",
      "1000 input query terms found no hit:\t['cg08130643', 'cg08130663', 'cg08130668', 'cg08130682', 'cg08130783', 'cg08130793', 'cg08130813', '\n",
      "1000 input query terms found no hit:\t['cg08183341', 'cg08183407', 'cg08183425', 'cg08183490', 'cg08183577', 'cg08183663', 'cg08183724', '\n",
      "1000 input query terms found no hit:\t['cg08238777', 'cg08238794', 'cg08239041', 'cg08239103', 'cg08239177', 'cg08239243', 'cg08239282', '\n",
      "1000 input query terms found no hit:\t['cg08289272', 'cg08289323', 'cg08289346', 'cg08289350', 'cg08289366', 'cg08289375', 'cg08289409', '\n",
      "1000 input query terms found no hit:\t['cg08347047', 'cg08347183', 'cg08347226', 'cg08347373', 'cg08347423', 'cg08347442', 'cg08347471', '\n",
      "1000 input query terms found no hit:\t['cg08409165', 'cg08409167', 'cg08409173', 'cg08409225', 'cg08409301', 'cg08409335', 'cg08409361', '\n",
      "1000 input query terms found no hit:\t['cg08465579', 'cg08465708', 'cg08465774', 'cg08465785', 'cg08465862', 'cg08466030', 'cg08466034', '\n",
      "1000 input query terms found no hit:\t['cg08522676', 'cg08522707', 'cg08522784', 'cg08522831', 'cg08523029', 'cg08523075', 'cg08523258', '\n",
      "1000 input query terms found no hit:\t['cg08578505', 'cg08578510', 'cg08578520', 'cg08578568', 'cg08578641', 'cg08578703', 'cg08578734', '\n",
      "1000 input query terms found no hit:\t['cg08636203', 'cg08636224', 'cg08636385', 'cg08636536', 'cg08636573', 'cg08636620', 'cg08636638', '\n",
      "1000 input query terms found no hit:\t['cg08694618', 'cg08694680', 'cg08694699', 'cg08694723', 'cg08694735', 'cg08694802', 'cg08694858', '\n",
      "1000 input query terms found no hit:\t['cg08751472', 'cg08751499', 'cg08751508', 'cg08751734', 'cg08751840', 'cg08751845', 'cg08751876', '\n",
      "1000 input query terms found no hit:\t['cg08814268', 'cg08814373', 'cg08814381', 'cg08814531', 'cg08814536', 'cg08814640', 'cg08814698', '\n",
      "1000 input query terms found no hit:\t['cg08864944', 'cg08865099', 'cg08865208', 'cg08865350', 'cg08865365', 'cg08865423', 'cg08865458', '\n",
      "1000 input query terms found no hit:\t['cg08918057', 'cg08918120', 'cg08918206', 'cg08918224', 'cg08918272', 'cg08918274', 'cg08918383', '\n",
      "1000 input query terms found no hit:\t['cg08977536', 'cg08977611', 'cg08977627', 'cg08977639', 'cg08977702', 'cg08977790', 'cg08977791', '\n",
      "1000 input query terms found no hit:\t['cg09041183', 'cg09041207', 'cg09041251', 'cg09041268', 'cg09041337', 'cg09041389', 'cg09041483', '\n",
      "1000 input query terms found no hit:\t['cg09105442', 'cg09105479', 'cg09105523', 'cg09105657', 'cg09105683', 'cg09105687', 'cg09105802', '\n",
      "1000 input query terms found no hit:\t['cg09161592', 'cg09161758', 'cg09162056', 'cg09162137', 'cg09162146', 'cg09162148', 'cg09162158', '\n",
      "1000 input query terms found no hit:\t['cg09225840', 'cg09225861', 'cg09225922', 'cg09225948', 'cg09225973', 'cg09226044', 'cg09226051', '\n",
      "1000 input query terms found no hit:\t['cg09277365', 'cg09277376', 'cg09277467', 'cg09277546', 'cg09277566', 'cg09277575', 'cg09277709', '\n",
      "1000 input query terms found no hit:\t['cg09333532', 'cg09333580', 'cg09333584', 'cg09333631', 'cg09333723', 'cg09333785', 'cg09333812', '\n",
      "1000 input query terms found no hit:\t['cg09394681', 'cg09394785', 'cg09395021', 'cg09395033', 'cg09395034', 'cg09395085', 'cg09395187', '\n",
      "1000 input query terms found no hit:\t['cg09450097', 'cg09450153', 'cg09450183', 'cg09450197', 'cg09450200', 'cg09450260', 'cg09450352', '\n",
      "1000 input query terms found no hit:\t['cg09508770', 'cg09508934', 'cg09508938', 'cg09509035', 'cg09509089', 'cg09509144', 'cg09509179', '\n",
      "1000 input query terms found no hit:\t['cg09565404', 'cg09565472', 'cg09565526', 'cg09565576', 'cg09565597', 'cg09565639', 'cg09565657', '\n",
      "1000 input query terms found no hit:\t['cg09627060', 'cg09627192', 'cg09627207', 'cg09627247', 'cg09627276', 'cg09627302', 'cg09627339', '\n",
      "1000 input query terms found no hit:\t['cg09685630', 'cg09685682', 'cg09685705', 'cg09685747', 'cg09685763', 'cg09685934', 'cg09685982', '\n",
      "1000 input query terms found no hit:\t['cg09751728', 'cg09751810', 'cg09751818', 'cg09751842', 'cg09752011', 'cg09752021', 'cg09752199', '\n",
      "1000 input query terms found no hit:\t['cg09809932', 'cg09809950', 'cg09810059', 'cg09810078', 'cg09810089', 'cg09810169', 'cg09810313', '\n",
      "1000 input query terms found no hit:\t['cg09864712', 'cg09864715', 'cg09864843', 'cg09864858', 'cg09864961', 'cg09864990', 'cg09864992', '\n",
      "1000 input query terms found no hit:\t['cg09923456', 'cg09923593', 'cg09923656', 'cg09923671', 'cg09923756', 'cg09923769', 'cg09923855', '\n",
      "1000 input query terms found no hit:\t['cg09987663', 'cg09987740', 'cg09987789', 'cg09987889', 'cg09987993', 'cg09988059', 'cg09988062', '\n",
      "1000 input query terms found no hit:\t['cg10046227', 'cg10046388', 'cg10046401', 'cg10046451', 'cg10046620', 'cg10046671', 'cg10046767', '\n",
      "1000 input query terms found no hit:\t['cg10101600', 'cg10101634', 'cg10101680', 'cg10101749', 'cg10101773', 'cg10101783', 'cg10101862', '\n",
      "1000 input query terms found no hit:\t['cg10154973', 'cg10155026', 'cg10155101', 'cg10155126', 'cg10155147', 'cg10155200', 'cg10155205', '\n",
      "1000 input query terms found no hit:\t['cg10207530', 'cg10207553', 'cg10207609', 'cg10207631', 'cg10207656', 'cg10207662', 'cg10207745', '\n",
      "1000 input query terms found no hit:\t['cg10266188', 'cg10266189', 'cg10266196', 'cg10266211', 'cg10266336', 'cg10266386', 'cg10266394', '\n",
      "1000 input query terms found no hit:\t['cg10325668', 'cg10325694', 'cg10326031', 'cg10326182', 'cg10326224', 'cg10326234', 'cg10326287', '\n",
      "1000 input query terms found no hit:\t['cg10384570', 'cg10384725', 'cg10384784', 'cg10384919', 'cg10385089', 'cg10385101', 'cg10385188', '\n",
      "1000 input query terms found no hit:\t['cg10448562', 'cg10448608', 'cg10448687', 'cg10448692', 'cg10448719', 'cg10448724', 'cg10448759', '\n",
      "1000 input query terms found no hit:\t['cg10505443', 'cg10505580', 'cg10505610', 'cg10505630', 'cg10505658', 'cg10505718', 'cg10505740', '\n",
      "1000 input query terms found no hit:\t['cg10558063', 'cg10558228', 'cg10558233', 'cg10558494', 'cg10558598', 'cg10558616', 'cg10558740', '\n",
      "1000 input query terms found no hit:\t['cg10617396', 'cg10617494', 'cg10617739', 'cg10617782', 'cg10617784', 'cg10617909', 'cg10617943', '\n",
      "1000 input query terms found no hit:\t['cg10677909', 'cg10677933', 'cg10677959', 'cg10678018', 'cg10678071', 'cg10678190', 'cg10678215', '\n",
      "1000 input query terms found no hit:\t['cg10744196', 'cg10744200', 'cg10744248', 'cg10744268', 'cg10744297', 'cg10744305', 'cg10744424', '\n",
      "1000 input query terms found no hit:\t['cg10803903', 'cg10803923', 'cg10803935', 'cg10804085', 'cg10804090', 'cg10804254', 'cg10804344', '\n",
      "1000 input query terms found no hit:\t['cg10862399', 'cg10862468', 'cg10862489', 'cg10862535', 'cg10862567', 'cg10862587', 'cg10862603', '\n",
      "1000 input query terms found no hit:\t['cg10929442', 'cg10929582', 'cg10929614', 'cg10929628', 'cg10929690', 'cg10929758', 'cg10929761', '\n",
      "1000 input query terms found no hit:\t['cg10987992', 'cg10988027', 'cg10988041', 'cg10988075', 'cg10988124', 'cg10988144', 'cg10988164', '\n",
      "1000 input query terms found no hit:\t['cg11051738', 'cg11051752', 'cg11051843', 'cg11051893', 'cg11051899', 'cg11052068', 'cg11052081', '\n",
      "1000 input query terms found no hit:\t['cg11115427', 'cg11115431', 'cg11115502', 'cg11115577', 'cg11115582', 'cg11115622', 'cg11115673', '\n",
      "1000 input query terms found no hit:\t['cg11177552', 'cg11177614', 'cg11177693', 'cg11177740', 'cg11177833', 'cg11177858', 'cg11177955', '\n",
      "1000 input query terms found no hit:\t['cg11231659', 'cg11231701', 'cg11231715', 'cg11231735', 'cg11231740', 'cg11231751', 'cg11231849', '\n",
      "1000 input query terms found no hit:\t['cg11293821', 'cg11293828', 'cg11294011', 'cg11294084', 'cg11294241', 'cg11294269', 'cg11294283', '\n",
      "1000 input query terms found no hit:\t['cg11353368', 'cg11353380', 'cg11353547', 'cg11353548', 'cg11353593', 'cg11353598', 'cg11353647', '\n",
      "1000 input query terms found no hit:\t['cg11414263', 'cg11414267', 'cg11414276', 'cg11414345', 'cg11414415', 'cg11414505', 'cg11414507', '\n",
      "1000 input query terms found no hit:\t['cg11484348', 'cg11484353', 'cg11484413', 'cg11484543', 'cg11484553', 'cg11484576', 'cg11484721', '\n",
      "1000 input query terms found no hit:\t['cg11552078', 'cg11552079', 'cg11552183', 'cg11552287', 'cg11552293', 'cg11552363', 'cg11552585', '\n",
      "1000 input query terms found no hit:\t['cg11613545', 'cg11613559', 'cg11613735', 'cg11613741', 'cg11613875', 'cg11613890', 'cg11613902', '\n",
      "1000 input query terms found no hit:\t['cg11673505', 'cg11673687', 'cg11673741', 'cg11673833', 'cg11673840', 'cg11673920', 'cg11673964', '\n",
      "1000 input query terms found no hit:\t['cg11731890', 'cg11731892', 'cg11731939', 'cg11732134', 'cg11732197', 'cg11732226', 'cg11732257', '\n",
      "1000 input query terms found no hit:\t['cg11791770', 'cg11791847', 'cg11791914', 'cg11791921', 'cg11791922', 'cg11791960', 'cg11792075', '\n",
      "1000 input query terms found no hit:\t['cg11842450', 'cg11842502', 'cg11842610', 'cg11842752', 'cg11842791', 'cg11842944', 'cg11842953', '\n",
      "1000 input query terms found no hit:\t['cg11899596', 'cg11899622', 'cg11899891', 'cg11899895', 'cg11900120', 'cg11900210', 'cg11900328', '\n",
      "1000 input query terms found no hit:\t['cg11960115', 'cg11960229', 'cg11960243', 'cg11960377', 'cg11960393', 'cg11960435', 'cg11960592', '\n",
      "1000 input query terms found no hit:\t['cg12026625', 'cg12026629', 'cg12026729', 'cg12026749', 'cg12026858', 'cg12026891', 'cg12026956', '\n",
      "1000 input query terms found no hit:\t['cg12078857', 'cg12078872', 'cg12078913', 'cg12078929', 'cg12078933', 'cg12078937', 'cg12078946', '\n",
      "1000 input query terms found no hit:\t['cg12138124', 'cg12138150', 'cg12138286', 'cg12138330', 'cg12138346', 'cg12138450', 'cg12138483', '\n",
      "1000 input query terms found no hit:\t['cg12198495', 'cg12198605', 'cg12198633', 'cg12198639', 'cg12198704', 'cg12198729', 'cg12198769', '\n",
      "1000 input query terms found no hit:\t['cg12263990', 'cg12264049', 'cg12264333', 'cg12264551', 'cg12264626', 'cg12264934', 'cg12264949', '\n",
      "1000 input query terms found no hit:\t['cg12338097', 'cg12338127', 'cg12338137', 'cg12338176', 'cg12338202', 'cg12338214', 'cg12338273', '\n",
      "1000 input query terms found no hit:\t['cg12406726', 'cg12406822', 'cg12406825', 'cg12406875', 'cg12406879', 'cg12406941', 'cg12406992', '\n",
      "1000 input query terms found no hit:\t['cg12466919', 'cg12466950', 'cg12467096', 'cg12467123', 'cg12467140', 'cg12467143', 'cg12467191', '\n",
      "1000 input query terms found no hit:\t['cg12528144', 'cg12528229', 'cg12528349', 'cg12528394', 'cg12528397', 'cg12528403', 'cg12528452', '\n",
      "1000 input query terms found no hit:\t['cg12591791', 'cg12591951', 'cg12591952', 'cg12592179', 'cg12592194', 'cg12592276', 'cg12592321', '\n",
      "1000 input query terms found no hit:\t['cg12647861', 'cg12647895', 'cg12647920', 'cg12647932', 'cg12647952', 'cg12647962', 'cg12647966', '\n",
      "1000 input query terms found no hit:\t['cg12715741', 'cg12715759', 'cg12715811', 'cg12715979', 'cg12716083', 'cg12716232', 'cg12716346', '\n",
      "1000 input query terms found no hit:\t['cg12787036', 'cg12787072', 'cg12787209', 'cg12787292', 'cg12787323', 'cg12787405', 'cg12787407', '\n",
      "1000 input query terms found no hit:\t['cg12850892', 'cg12850990', 'cg12851049', 'cg12851161', 'cg12851190', 'cg12851401', 'cg12851429', '\n",
      "1000 input query terms found no hit:\t['cg12918604', 'cg12918696', 'cg12918895', 'cg12918961', 'cg12919006', 'cg12919025', 'cg12919119', '\n",
      "1000 input query terms found no hit:\t['cg12986611', 'cg12986628', 'cg12986639', 'cg12986659', 'cg12986700', 'cg12986726', 'cg12986955', '\n",
      "1000 input query terms found no hit:\t['cg13055450', 'cg13055484', 'cg13055665', 'cg13055685', 'cg13055709', 'cg13056222', 'cg13056247', '\n",
      "1000 input query terms found no hit:\t['cg13124863', 'cg13125072', 'cg13125157', 'cg13125224', 'cg13125364', 'cg13125488', 'cg13125506', '\n",
      "1000 input query terms found no hit:\t['cg13190531', 'cg13190540', 'cg13190572', 'cg13190608', 'cg13190879', 'cg13190924', 'cg13190986', '\n",
      "1000 input query terms found no hit:\t['cg13253303', 'cg13253439', 'cg13253481', 'cg13253637', 'cg13253642', 'cg13253660', 'cg13253694', '\n",
      "1000 input query terms found no hit:\t['cg13316719', 'cg13316720', 'cg13316736', 'cg13316813', 'cg13316854', 'cg13316864', 'cg13316877', '\n",
      "1000 input query terms found no hit:\t['cg13375255', 'cg13375262', 'cg13375295', 'cg13375352', 'cg13375427', 'cg13375457', 'cg13375462', '\n",
      "1000 input query terms found no hit:\t['cg13426857', 'cg13426876', 'cg13426891', 'cg13427055', 'cg13427149', 'cg13427179', 'cg13427189', '\n",
      "1000 input query terms found no hit:\t['cg13470462', 'cg13470477', 'cg13470499', 'cg13470504', 'cg13470557', 'cg13470644', 'cg13470658', '\n",
      "1000 input query terms found no hit:\t['cg13519464', 'cg13519536', 'cg13519549', 'cg13519595', 'cg13519696', 'cg13519851', 'cg13519863', '\n",
      "1000 input query terms found no hit:\t['cg13566278', 'cg13566279', 'cg13566430', 'cg13566436', 'cg13566468', 'cg13566596', 'cg13566610', '\n",
      "1000 input query terms found no hit:\t['cg13614045', 'cg13614060', 'cg13614083', 'cg13614092', 'cg13614103', 'cg13614181', 'cg13614229', '\n",
      "1000 input query terms found no hit:\t['cg13663978', 'cg13664067', 'cg13664079', 'cg13664102', 'cg13664146', 'cg13664345', 'cg13664579', '\n",
      "1000 input query terms found no hit:\t['cg13710969', 'cg13711063', 'cg13711129', 'cg13711144', 'cg13711186', 'cg13711197', 'cg13711238', '\n",
      "1000 input query terms found no hit:\t['cg13764350', 'cg13764516', 'cg13764652', 'cg13764778', 'cg13764828', 'cg13764850', 'cg13764897', '\n",
      "1000 input query terms found no hit:\t['cg13812452', 'cg13812570', 'cg13812587', 'cg13812613', 'cg13812731', 'cg13812879', 'cg13812894', '\n",
      "1000 input query terms found no hit:\t['cg13866767', 'cg13866795', 'cg13866910', 'cg13866977', 'cg13867039', 'cg13867237', 'cg13867248', '\n",
      "1000 input query terms found no hit:\t['cg13916500', 'cg13916516', 'cg13916603', 'cg13916633', 'cg13916740', 'cg13916742', 'cg13916816', '\n",
      "1000 input query terms found no hit:\t['cg13965272', 'cg13965276', 'cg13965321', 'cg13965328', 'cg13965612', 'cg13965707', 'cg13965724', '\n",
      "1000 input query terms found no hit:\t['cg14011845', 'cg14011854', 'cg14011971', 'cg14011987', 'cg14012014', 'cg14012033', 'cg14012082', '\n",
      "1000 input query terms found no hit:\t['cg14058848', 'cg14058851', 'cg14058861', 'cg14058863', 'cg14058869', 'cg14058932', 'cg14058998', '\n",
      "1000 input query terms found no hit:\t['cg14107256', 'cg14107273', 'cg14107328', 'cg14107346', 'cg14107382', 'cg14107457', 'cg14107488', '\n",
      "1000 input query terms found no hit:\t['cg14160433', 'cg14160449', 'cg14160480', 'cg14160518', 'cg14160556', 'cg14160599', 'cg14160632', '\n",
      "1000 input query terms found no hit:\t['cg14206172', 'cg14206273', 'cg14206365', 'cg14206479', 'cg14206510', 'cg14206516', 'cg14206523', '\n",
      "1000 input query terms found no hit:\t['cg14252164', 'cg14252169', 'cg14252211', 'cg14252222', 'cg14252236', 'cg14252237', 'cg14252279', '\n",
      "1000 input query terms found no hit:\t['cg14301425', 'cg14301531', 'cg14301562', 'cg14301635', 'cg14301710', 'cg14301713', 'cg14301764', '\n",
      "1000 input query terms found no hit:\t['cg14353201', 'cg14353208', 'cg14353218', 'cg14353231', 'cg14353258', 'cg14353278', 'cg14353373', '\n",
      "1000 input query terms found no hit:\t['cg14408356', 'cg14408412', 'cg14408416', 'cg14408428', 'cg14408431', 'cg14408447', 'cg14408465', '\n",
      "1000 input query terms found no hit:\t['cg14459668', 'cg14459710', 'cg14459716', 'cg14459772', 'cg14459960', 'cg14459995', 'cg14460003', '\n",
      "1000 input query terms found no hit:\t['cg14509606', 'cg14509623', 'cg14509631', 'cg14509761', 'cg14509802', 'cg14509809', 'cg14509895', '\n",
      "1000 input query terms found no hit:\t['cg14554716', 'cg14554743', 'cg14554776', 'cg14554788', 'cg14554813', 'cg14554846', 'cg14554869', '\n",
      "1000 input query terms found no hit:\t['cg14601938', 'cg14601981', 'cg14602087', 'cg14602164', 'cg14602220', 'cg14602222', 'cg14602234', '\n",
      "1000 input query terms found no hit:\t['cg14654468', 'cg14654471', 'cg14654495', 'cg14654533', 'cg14654587', 'cg14654613', 'cg14654629', '\n",
      "1000 input query terms found no hit:\t['cg14713139', 'cg14713146', 'cg14713153', 'cg14713217', 'cg14713231', 'cg14713449', 'cg14713933', '\n",
      "1000 input query terms found no hit:\t['cg14778437', 'cg14778531', 'cg14778560', 'cg14778576', 'cg14778608', 'cg14778622', 'cg14778644', '\n",
      "1000 input query terms found no hit:\t['cg14837792', 'cg14837800', 'cg14837817', 'cg14837825', 'cg14837848', 'cg14837896', 'cg14837919', '\n",
      "1000 input query terms found no hit:\t['cg14898177', 'cg14898223', 'cg14898232', 'cg14898243', 'cg14898260', 'cg14898263', 'cg14898332', '\n",
      "1000 input query terms found no hit:\t['cg14962049', 'cg14962120', 'cg14962136', 'cg14962158', 'cg14962159', 'cg14962197', 'cg14962327', '\n",
      "1000 input query terms found no hit:\t['cg15020894', 'cg15020964', 'cg15021031', 'cg15021224', 'cg15021238', 'cg15021280', 'cg15021292', '\n",
      "1000 input query terms found no hit:\t['cg15080403', 'cg15080430', 'cg15080433', 'cg15080590', 'cg15080701', 'cg15080754', 'cg15080866', '\n",
      "1000 input query terms found no hit:\t['cg15138382', 'cg15138390', 'cg15138396', 'cg15138473', 'cg15138543', 'cg15138570', 'cg15138614', '\n",
      "1000 input query terms found no hit:\t['cg15205559', 'cg15205568', 'cg15205636', 'cg15205733', 'cg15206062', 'cg15206136', 'cg15206171', '\n",
      "1000 input query terms found no hit:\t['cg15275286', 'cg15275309', 'cg15275312', 'cg15275313', 'cg15275393', 'cg15275427', 'cg15275625', '\n",
      "1000 input query terms found no hit:\t['cg15350780', 'cg15350840', 'cg15350899', 'cg15350925', 'cg15351009', 'cg15351061', 'cg15351067', '\n",
      "1000 input query terms found no hit:\t['cg15416233', 'cg15416250', 'cg15416279', 'cg15416329', 'cg15416352', 'cg15416363', 'cg15416417', '\n",
      "1000 input query terms found no hit:\t['cg15475851', 'cg15475967', 'cg15475971', 'cg15475977', 'cg15476227', 'cg15476229', 'cg15476279', '\n",
      "1000 input query terms found no hit:\t['cg15548656', 'cg15548711', 'cg15548859', 'cg15549029', 'cg15549075', 'cg15549100', 'cg15549126', '\n",
      "1000 input query terms found no hit:\t['cg15606393', 'cg15606473', 'cg15606604', 'cg15606633', 'cg15606654', 'cg15606663', 'cg15606692', '\n",
      "1000 input query terms found no hit:\t['cg15663092', 'cg15663121', 'cg15663173', 'cg15663265', 'cg15663316', 'cg15663348', 'cg15663484', '\n",
      "1000 input query terms found no hit:\t['cg15721142', 'cg15721145', 'cg15721238', 'cg15721243', 'cg15721359', 'cg15721424', 'cg15721439', '\n",
      "1000 input query terms found no hit:\t['cg15787807', 'cg15787985', 'cg15788147', 'cg15788149', 'cg15788231', 'cg15788265', 'cg15788279', '\n",
      "1000 input query terms found no hit:\t['cg15847604', 'cg15847614', 'cg15847685', 'cg15847775', 'cg15847845', 'cg15847896', 'cg15847926', '\n",
      "1000 input query terms found no hit:\t['cg15910994', 'cg15911114', 'cg15911153', 'cg15911165', 'cg15911239', 'cg15911328', 'cg15911409', '\n",
      "1000 input query terms found no hit:\t['cg15977272', 'cg15977278', 'cg15977432', 'cg15977634', 'cg15977661', 'cg15977685', 'cg15977716', '\n",
      "1000 input query terms found no hit:\t['cg16033053', 'cg16033151', 'cg16033274', 'cg16033277', 'cg16033299', 'cg16033376', 'cg16033410', '\n",
      "1000 input query terms found no hit:\t['cg16099353', 'cg16099404', 'cg16099496', 'cg16099599', 'cg16099624', 'cg16099687', 'cg16099707', '\n",
      "1000 input query terms found no hit:\t['cg16163981', 'cg16164062', 'cg16164079', 'cg16164092', 'cg16164144', 'cg16164276', 'cg16164356', '\n",
      "1000 input query terms found no hit:\t['cg16222840', 'cg16222848', 'cg16222850', 'cg16222896', 'cg16223079', 'cg16223220', 'cg16223358', '\n",
      "1000 input query terms found no hit:\t['cg16279775', 'cg16279786', 'cg16279835', 'cg16279861', 'cg16279909', 'cg16279940', 'cg16279999', '\n",
      "1000 input query terms found no hit:\t['cg16336752', 'cg16336802', 'cg16336872', 'cg16336989', 'cg16337087', 'cg16337108', 'cg16337138', '\n",
      "1000 input query terms found no hit:\t['cg16396284', 'cg16396309', 'cg16396396', 'cg16396414', 'cg16396417', 'cg16396441', 'cg16396488', '\n",
      "1000 input query terms found no hit:\t['cg16449402', 'cg16449464', 'cg16449476', 'cg16449489', 'cg16449606', 'cg16449620', 'cg16449636', '\n",
      "1000 input query terms found no hit:\t['cg16511870', 'cg16511948', 'cg16511983', 'cg16511985', 'cg16511994', 'cg16512060', 'cg16512069', '\n",
      "1000 input query terms found no hit:\t['cg16563370', 'cg16563454', 'cg16563470', 'cg16563510', 'cg16563569', 'cg16563656', 'cg16563660', '\n",
      "1000 input query terms found no hit:\t['cg16616993', 'cg16617073', 'cg16617097', 'cg16617141', 'cg16617243', 'cg16617257', 'cg16617297', '\n",
      "1000 input query terms found no hit:\t['cg16670736', 'cg16670809', 'cg16670864', 'cg16670889', 'cg16670940', 'cg16671069', 'cg16671079', '\n",
      "1000 input query terms found no hit:\t['cg16721101', 'cg16721106', 'cg16721177', 'cg16721191', 'cg16721194', 'cg16721202', 'cg16721257', '\n",
      "1000 input query terms found no hit:\t['cg16783204', 'cg16783279', 'cg16783300', 'cg16783314', 'cg16783349', 'cg16783412', 'cg16783478', '\n",
      "1000 input query terms found no hit:\t['cg16851359', 'cg16851385', 'cg16851402', 'cg16851417', 'cg16851423', 'cg16851425', 'cg16851482', '\n",
      "1000 input query terms found no hit:\t['cg16915828', 'cg16915863', 'cg16915916', 'cg16916165', 'cg16916522', 'cg16916549', 'cg16916688', '\n",
      "1000 input query terms found no hit:\t['cg16989544', 'cg16989587', 'cg16989646', 'cg16989714', 'cg16989719', 'cg16989784', 'cg16989797', '\n",
      "1000 input query terms found no hit:\t['cg17051591', 'cg17051623', 'cg17051628', 'cg17051704', 'cg17051733', 'cg17051905', 'cg17051995', '\n",
      "1000 input query terms found no hit:\t['cg17114847', 'cg17114866', 'cg17114932', 'cg17115147', 'cg17115172', 'cg17115250', 'cg17115258', '\n",
      "1000 input query terms found no hit:\t['cg17174493', 'cg17174566', 'cg17174613', 'cg17174660', 'cg17174775', 'cg17174778', 'cg17174814', '\n",
      "1000 input query terms found no hit:\t['cg17237086', 'cg17237111', 'cg17237181', 'cg17237320', 'cg17237460', 'cg17237506', 'cg17237603', '\n",
      "1000 input query terms found no hit:\t['cg17299935', 'cg17299957', 'cg17300047', 'cg17300051', 'cg17300237', 'cg17300275', 'cg17300307', '\n",
      "1000 input query terms found no hit:\t['cg17358307', 'cg17358422', 'cg17358530', 'cg17358883', 'cg17358899', 'cg17359064', 'cg17359076', '\n",
      "1000 input query terms found no hit:\t['cg17422427', 'cg17422460', 'cg17422462', 'cg17422500', 'cg17422543', 'cg17422603', 'cg17422704', '\n",
      "1000 input query terms found no hit:\t['cg17482089', 'cg17482114', 'cg17482117', 'cg17482197', 'cg17482224', 'cg17482237', 'cg17482273', '\n",
      "1000 input query terms found no hit:\t['cg17542594', 'cg17542650', 'cg17542751', 'cg17542795', 'cg17542806', 'cg17542835', 'cg17542852', '\n",
      "1000 input query terms found no hit:\t['cg17615115', 'cg17615629', 'cg17615682', 'cg17615744', 'cg17615773', 'cg17616099', 'cg17616192', '\n",
      "1000 input query terms found no hit:\t['cg17694795', 'cg17694848', 'cg17694851', 'cg17694877', 'cg17695010', 'cg17695145', 'cg17695351', '\n",
      "1000 input query terms found no hit:\t['cg17762073', 'cg17762124', 'cg17762146', 'cg17762168', 'cg17762328', 'cg17762331', 'cg17762420', '\n",
      "1000 input query terms found no hit:\t['cg17826786', 'cg17826834', 'cg17826893', 'cg17826956', 'cg17826980', 'cg17826986', 'cg17826989', '\n",
      "1000 input query terms found no hit:\t['cg17884201', 'cg17884294', 'cg17884327', 'cg17884373', 'cg17884383', 'cg17884674', 'cg17884698', '\n",
      "1000 input query terms found no hit:\t['cg17951613', 'cg17951713', 'cg17951850', 'cg17951878', 'cg17951978', 'cg17951991', 'cg17952046', '\n",
      "1000 input query terms found no hit:\t['cg18015534', 'cg18015622', 'cg18015625', 'cg18015677', 'cg18015704', 'cg18015746', 'cg18015787', '\n",
      "1000 input query terms found no hit:\t['cg18075040', 'cg18075073', 'cg18075115', 'cg18075117', 'cg18075287', 'cg18075299', 'cg18075315', '\n",
      "1000 input query terms found no hit:\t['cg18127934', 'cg18127957', 'cg18127966', 'cg18128058', 'cg18128147', 'cg18128164', 'cg18128212', '\n",
      "1000 input query terms found no hit:\t['cg18184880', 'cg18184910', 'cg18184935', 'cg18184941', 'cg18184975', 'cg18185008', 'cg18185028', '\n",
      "1000 input query terms found no hit:\t['cg18245268', 'cg18245281', 'cg18245306', 'cg18245316', 'cg18245365', 'cg18245460', 'cg18245631', '\n",
      "1000 input query terms found no hit:\t['cg18311498', 'cg18311516', 'cg18311537', 'cg18311633', 'cg18311665', 'cg18311708', 'cg18311854', '\n",
      "1000 input query terms found no hit:\t['cg18376227', 'cg18376256', 'cg18376288', 'cg18376306', 'cg18376335', 'cg18376352', 'cg18376472', '\n",
      "1000 input query terms found no hit:\t['cg18436203', 'cg18436214', 'cg18436246', 'cg18436272', 'cg18436324', 'cg18436395', 'cg18436444', '\n",
      "1000 input query terms found no hit:\t['cg18494018', 'cg18494043', 'cg18494124', 'cg18494188', 'cg18494190', 'cg18494192', 'cg18494272', '\n",
      "1000 input query terms found no hit:\t['cg18558647', 'cg18558744', 'cg18558763', 'cg18558767', 'cg18558770', 'cg18558808', 'cg18558811', '\n",
      "1000 input query terms found no hit:\t['cg18622954', 'cg18623216', 'cg18623263', 'cg18623351', 'cg18623355', 'cg18623476', 'cg18623649', '\n",
      "1000 input query terms found no hit:\t['cg18692382', 'cg18692507', 'cg18692596', 'cg18692678', 'cg18692705', 'cg18692955', 'cg18693004', '\n",
      "1000 input query terms found no hit:\t['cg18755829', 'cg18755846', 'cg18755855', 'cg18755942', 'cg18755950', 'cg18755984', 'cg18756060', '\n",
      "1000 input query terms found no hit:\t['cg18811601', 'cg18811621', 'cg18811731', 'cg18811789', 'cg18811819', 'cg18811860', 'cg18811916', '\n",
      "1000 input query terms found no hit:\t['cg18874425', 'cg18874502', 'cg18874575', 'cg18874902', 'cg18874907', 'cg18874953', 'cg18875012', '\n",
      "1000 input query terms found no hit:\t['cg18942451', 'cg18942464', 'cg18942579', 'cg18942631', 'cg18942715', 'cg18942757', 'cg18942813', '\n",
      "1000 input query terms found no hit:\t['cg19010441', 'cg19010490', 'cg19010528', 'cg19010566', 'cg19010777', 'cg19010846', 'cg19010850', '\n",
      "1000 input query terms found no hit:\t['cg19082496', 'cg19082559', 'cg19082560', 'cg19082708', 'cg19082888', 'cg19082909', 'cg19082920', '\n",
      "1000 input query terms found no hit:\t['cg19150666', 'cg19150842', 'cg19150852', 'cg19150884', 'cg19150916', 'cg19150921', 'cg19151030', '\n",
      "1000 input query terms found no hit:\t['cg19227170', 'cg19227382', 'cg19227485', 'cg19227511', 'cg19227523', 'cg19227561', 'cg19227606', '\n",
      "1000 input query terms found no hit:\t['cg19287349', 'cg19287501', 'cg19287591', 'cg19287610', 'cg19287711', 'cg19287714', 'cg19287808', '\n",
      "1000 input query terms found no hit:\t['cg19360104', 'cg19360212', 'cg19360308', 'cg19360316', 'cg19360330', 'cg19360425', 'cg19360482', '\n",
      "1000 input query terms found no hit:\t['cg19422253', 'cg19422300', 'cg19422304', 'cg19422340', 'cg19422369', 'cg19422460', 'cg19422565', '\n",
      "1000 input query terms found no hit:\t['cg19486833', 'cg19486949', 'cg19486955', 'cg19487455', 'cg19487484', 'cg19487717', 'cg19487718', '\n",
      "1000 input query terms found no hit:\t['cg19547844', 'cg19547972', 'cg19547976', 'cg19548004', 'cg19548119', 'cg19548176', 'cg19548181', '\n",
      "1000 input query terms found no hit:\t['cg19609502', 'cg19609644', 'cg19609912', 'cg19609923', 'cg19610126', 'cg19610177', 'cg19610253', '\n",
      "1000 input query terms found no hit:\t['cg19673995', 'cg19674032', 'cg19674051', 'cg19674065', 'cg19674091', 'cg19674138', 'cg19674166', '\n",
      "1000 input query terms found no hit:\t['cg19730314', 'cg19730379', 'cg19730600', 'cg19730687', 'cg19730691', 'cg19730699', 'cg19730798', '\n",
      "1000 input query terms found no hit:\t['cg19782746', 'cg19782749', 'cg19782880', 'cg19782923', 'cg19783015', 'cg19783114', 'cg19783150', '\n",
      "1000 input query terms found no hit:\t['cg19844333', 'cg19844581', 'cg19844635', 'cg19844653', 'cg19844704', 'cg19844829', 'cg19844840', '\n",
      "1000 input query terms found no hit:\t['cg19903298', 'cg19903333', 'cg19903596', 'cg19903702', 'cg19903738', 'cg19903753', 'cg19903766', '\n",
      "1000 input query terms found no hit:\t['cg19959591', 'cg19959621', 'cg19959647', 'cg19959648', 'cg19959677', 'cg19959917', 'cg19959974', '\n",
      "1000 input query terms found no hit:\t['cg20025135', 'cg20025389', 'cg20025403', 'cg20025630', 'cg20025656', 'cg20025658', 'cg20025677', '\n",
      "1000 input query terms found no hit:\t['cg20093940', 'cg20093992', 'cg20094066', 'cg20094085', 'cg20094111', 'cg20094279', 'cg20094282', '\n",
      "1000 input query terms found no hit:\t['cg20158669', 'cg20158671', 'cg20158796', 'cg20158826', 'cg20158867', 'cg20158944', 'cg20159036', '\n",
      "1000 input query terms found no hit:\t['cg20234855', 'cg20234910', 'cg20234947', 'cg20234959', 'cg20234976', 'cg20235051', 'cg20235075', '\n",
      "1000 input query terms found no hit:\t['cg20299703', 'cg20299740', 'cg20299776', 'cg20299802', 'cg20299810', 'cg20299935', 'cg20300032', '\n",
      "1000 input query terms found no hit:\t['cg20367218', 'cg20367223', 'cg20367258', 'cg20367266', 'cg20367304', 'cg20367329', 'cg20367388', '\n",
      "1000 input query terms found no hit:\t['cg20435623', 'cg20435725', 'cg20435896', 'cg20436037', 'cg20436086', 'cg20436206', 'cg20436262', '\n",
      "1000 input query terms found no hit:\t['cg20501958', 'cg20502003', 'cg20502006', 'cg20502039', 'cg20502080', 'cg20502215', 'cg20502257', '\n",
      "1000 input query terms found no hit:\t['cg20570410', 'cg20570458', 'cg20570464', 'cg20570505', 'cg20570544', 'cg20570611', 'cg20570682', '\n",
      "1000 input query terms found no hit:\t['cg20641095', 'cg20641280', 'cg20641289', 'cg20641355', 'cg20641423', 'cg20641450', 'cg20641465', '\n",
      "1000 input query terms found no hit:\t['cg20698781', 'cg20698862', 'cg20698924', 'cg20698942', 'cg20699036', 'cg20699076', 'cg20699097', '\n",
      "1000 input query terms found no hit:\t['cg20757997', 'cg20758098', 'cg20758219', 'cg20758492', 'cg20758542', 'cg20758593', 'cg20758651', '\n",
      "1000 input query terms found no hit:\t['cg20826277', 'cg20826345', 'cg20826348', 'cg20826379', 'cg20826416', 'cg20826442', 'cg20826526', '\n",
      "1000 input query terms found no hit:\t['cg20909380', 'cg20909579', 'cg20909645', 'cg20909656', 'cg20909664', 'cg20909686', 'cg20909773', '\n",
      "1000 input query terms found no hit:\t['cg20972834', 'cg20972917', 'cg20972950', 'cg20973129', 'cg20973210', 'cg20973260', 'cg20973347', '\n",
      "1000 input query terms found no hit:\t['cg21037784', 'cg21037892', 'cg21037935', 'cg21037943', 'cg21038123', 'cg21038156', 'cg21038220', '\n",
      "1000 input query terms found no hit:\t['cg21108767', 'cg21108826', 'cg21108841', 'cg21108851', 'cg21108859', 'cg21109013', 'cg21109025', '\n",
      "1000 input query terms found no hit:\t['cg21166964', 'cg21166985', 'cg21166989', 'cg21166999', 'cg21167046', 'cg21167072', 'cg21167106', '\n",
      "1000 input query terms found no hit:\t['cg21218561', 'cg21218627', 'cg21218636', 'cg21218646', 'cg21218687', 'cg21218820', 'cg21218848', '\n",
      "1000 input query terms found no hit:\t['cg21282630', 'cg21282663', 'cg21282669', 'cg21282736', 'cg21282837', 'cg21282883', 'cg21282907', '\n",
      "1000 input query terms found no hit:\t['cg21366200', 'cg21366321', 'cg21366535', 'cg21366602', 'cg21366673', 'cg21366688', 'cg21366834', '\n",
      "1000 input query terms found no hit:\t['cg21443274', 'cg21443284', 'cg21443584', 'cg21443659', 'cg21443699', 'cg21443763', 'cg21443773', '\n",
      "1000 input query terms found no hit:\t['cg21507636', 'cg21507694', 'cg21507719', 'cg21507807', 'cg21507854', 'cg21507866', 'cg21507882', '\n",
      "1000 input query terms found no hit:\t['cg21567504', 'cg21567625', 'cg21567649', 'cg21567958', 'cg21567971', 'cg21568081', 'cg21568106', '\n",
      "1000 input query terms found no hit:\t['cg21628854', 'cg21628927', 'cg21629006', 'cg21629166', 'cg21629325', 'cg21629384', 'cg21629394', '\n",
      "1000 input query terms found no hit:\t['cg21700723', 'cg21700772', 'cg21700774', 'cg21700780', 'cg21700831', 'cg21700863', 'cg21701043', '\n",
      "1000 input query terms found no hit:\t['cg21776286', 'cg21776403', 'cg21776417', 'cg21776419', 'cg21776524', 'cg21776568', 'cg21776577', '\n",
      "1000 input query terms found no hit:\t['cg21844331', 'cg21844450', 'cg21844472', 'cg21844656', 'cg21844669', 'cg21844722', 'cg21844726', '\n",
      "1000 input query terms found no hit:\t['cg21908178', 'cg21908208', 'cg21908235', 'cg21908248', 'cg21908259', 'cg21908287', 'cg21908344', '\n",
      "1000 input query terms found no hit:\t['cg21975834', 'cg21976560', 'cg21976603', 'cg21976687', 'cg21976880', 'cg21977057', 'cg21977075', '\n",
      "1000 input query terms found no hit:\t['cg22047901', 'cg22047910', 'cg22047977', 'cg22048132', 'cg22048216', 'cg22048228', 'cg22048274', '\n",
      "1000 input query terms found no hit:\t['cg22120488', 'cg22120509', 'cg22120738', 'cg22120825', 'cg22120852', 'cg22120894', 'cg22120948', '\n",
      "1000 input query terms found no hit:\t['cg22186216', 'cg22186223', 'cg22186263', 'cg22186291', 'cg22186315', 'cg22186380', 'cg22186503', '\n",
      "1000 input query terms found no hit:\t['cg22255773', 'cg22256027', 'cg22256180', 'cg22256295', 'cg22256309', 'cg22256354', 'cg22256433', '\n",
      "1000 input query terms found no hit:\t['cg22332266', 'cg22332276', 'cg22332306', 'cg22332321', 'cg22332339', 'cg22332388', 'cg22332577', '\n",
      "1000 input query terms found no hit:\t['cg22398959', 'cg22399023', 'cg22399096', 'cg22399111', 'cg22399133', 'cg22399232', 'cg22399423', '\n",
      "1000 input query terms found no hit:\t['cg22480109', 'cg22480459', 'cg22480490', 'cg22480555', 'cg22480558', 'cg22480742', 'cg22480783', '\n",
      "1000 input query terms found no hit:\t['cg22541400', 'cg22541509', 'cg22541520', 'cg22541535', 'cg22541679', 'cg22541735', 'cg22541755', '\n",
      "1000 input query terms found no hit:\t['cg22615515', 'cg22615730', 'cg22615747', 'cg22615819', 'cg22615910', 'cg22615917', 'cg22615948', '\n",
      "1000 input query terms found no hit:\t['cg22681998', 'cg22682031', 'cg22682154', 'cg22682161', 'cg22682162', 'cg22682172', 'cg22682200', '\n",
      "1000 input query terms found no hit:\t['cg22741264', 'cg22741272', 'cg22741595', 'cg22741626', 'cg22741653', 'cg22741735', 'cg22741823', '\n",
      "1000 input query terms found no hit:\t['cg22806527', 'cg22806557', 'cg22806587', 'cg22806837', 'cg22806907', 'cg22806908', 'cg22806934', '\n",
      "1000 input query terms found no hit:\t['cg22865720', 'cg22865769', 'cg22865798', 'cg22865824', 'cg22865831', 'cg22865905', 'cg22865908', '\n",
      "1000 input query terms found no hit:\t['cg22929010', 'cg22929018', 'cg22929099', 'cg22929104', 'cg22929219', 'cg22929235', 'cg22929280', '\n",
      "1000 input query terms found no hit:\t['cg22988237', 'cg22988346', 'cg22988430', 'cg22988444', 'cg22988458', 'cg22988483', 'cg22988507', '\n",
      "1000 input query terms found no hit:\t['cg23046918', 'cg23046919', 'cg23046934', 'cg23047077', 'cg23047079', 'cg23047134', 'cg23047137', '\n",
      "1000 input query terms found no hit:\t['cg23101091', 'cg23101093', 'cg23101118', 'cg23101138', 'cg23101259', 'cg23101464', 'cg23101469', '\n",
      "1000 input query terms found no hit:\t['cg23172480', 'cg23172545', 'cg23172579', 'cg23172606', 'cg23172664', 'cg23172671', 'cg23172828', '\n",
      "1000 input query terms found no hit:\t['cg23228745', 'cg23228779', 'cg23228835', 'cg23228858', 'cg23228887', 'cg23228921', 'cg23229016', '\n",
      "1000 input query terms found no hit:\t['cg23282025', 'cg23282051', 'cg23282276', 'cg23282363', 'cg23282441', 'cg23282442', 'cg23282480', '\n",
      "1000 input query terms found no hit:\t['cg23345697', 'cg23345864', 'cg23345930', 'cg23345951', 'cg23345977', 'cg23346001', 'cg23346107', '\n",
      "1000 input query terms found no hit:\t['cg23411013', 'cg23411036', 'cg23411043', 'cg23411065', 'cg23411103', 'cg23411121', 'cg23411150', '\n",
      "1000 input query terms found no hit:\t['cg23477774', 'cg23477849', 'cg23477940', 'cg23477967', 'cg23478098', 'cg23478105', 'cg23478124', '\n",
      "1000 input query terms found no hit:\t['cg23536681', 'cg23536830', 'cg23537141', 'cg23537193', 'cg23537305', 'cg23537415', 'cg23537419', '\n",
      "1000 input query terms found no hit:\t['cg23610820', 'cg23610841', 'cg23610994', 'cg23611097', 'cg23611220', 'cg23611233', 'cg23611477', '\n",
      "1000 input query terms found no hit:\t['cg23666299', 'cg23666300', 'cg23666362', 'cg23666374', 'cg23666378', 'cg23666457', 'cg23666491', '\n",
      "1000 input query terms found no hit:\t['cg23721301', 'cg23721345', 'cg23721398', 'cg23721425', 'cg23721533', 'cg23721586', 'cg23721598', '\n",
      "1000 input query terms found no hit:\t['cg23800169', 'cg23800202', 'cg23800248', 'cg23800435', 'cg23800437', 'cg23800573', 'cg23800587', '\n",
      "1000 input query terms found no hit:\t['cg23870168', 'cg23870181', 'cg23870282', 'cg23870378', 'cg23870587', 'cg23870762', 'cg23870769', '\n",
      "1000 input query terms found no hit:\t['cg23929628', 'cg23929674', 'cg23929682', 'cg23929809', 'cg23929834', 'cg23930083', 'cg23930114', '\n",
      "1000 input query terms found no hit:\t['cg23986219', 'cg23986335', 'cg23986375', 'cg23986376', 'cg23986385', 'cg23986470', 'cg23986474', '\n",
      "1000 input query terms found no hit:\t['cg24041977', 'cg24041995', 'cg24042127', 'cg24042242', 'cg24042275', 'cg24042313', 'cg24042452', '\n",
      "1000 input query terms found no hit:\t['cg24103651', 'cg24103799', 'cg24103837', 'cg24103979', 'cg24104045', 'cg24104099', 'cg24104128', '\n",
      "1000 input query terms found no hit:\t['cg24158448', 'cg24158452', 'cg24158482', 'cg24158506', 'cg24158546', 'cg24158553', 'cg24158594', '\n",
      "1000 input query terms found no hit:\t['cg24227390', 'cg24227397', 'cg24227481', 'cg24227590', 'cg24227728', 'cg24227730', 'cg24227837', '\n",
      "1000 input query terms found no hit:\t['cg24306277', 'cg24306278', 'cg24306340', 'cg24306353', 'cg24306397', 'cg24306422', 'cg24306430', '\n",
      "1000 input query terms found no hit:\t['cg24365117', 'cg24365121', 'cg24365142', 'cg24365215', 'cg24365417', 'cg24365464', 'cg24365518', '\n",
      "1000 input query terms found no hit:\t['cg24428760', 'cg24428851', 'cg24428877', 'cg24428913', 'cg24429037', 'cg24429081', 'cg24429159', '\n",
      "1000 input query terms found no hit:\t['cg24491576', 'cg24491618', 'cg24491704', 'cg24491749', 'cg24491766', 'cg24491776', 'cg24491784', '\n",
      "1000 input query terms found no hit:\t['cg24546622', 'cg24546680', 'cg24546683', 'cg24546723', 'cg24546928', 'cg24546942', 'cg24546984', '\n",
      "1000 input query terms found no hit:\t['cg24613524', 'cg24613558', 'cg24613644', 'cg24613691', 'cg24613798', 'cg24613906', 'cg24613923', '\n",
      "1000 input query terms found no hit:\t['cg24677444', 'cg24677472', 'cg24677501', 'cg24677671', 'cg24677674', 'cg24677732', 'cg24677744', '\n",
      "1000 input query terms found no hit:\t['cg24727561', 'cg24727568', 'cg24727620', 'cg24728005', 'cg24728008', 'cg24728018', 'cg24728050', '\n",
      "1000 input query terms found no hit:\t['cg24791188', 'cg24791218', 'cg24791283', 'cg24791360', 'cg24791380', 'cg24791428', 'cg24791546', '\n",
      "1000 input query terms found no hit:\t['cg24853724', 'cg24853765', 'cg24853868', 'cg24853871', 'cg24853956', 'cg24853975', 'cg24853978', '\n",
      "1000 input query terms found no hit:\t['cg24910112', 'cg24910161', 'cg24910352', 'cg24910374', 'cg24910410', 'cg24910675', 'cg24910739', '\n",
      "1000 input query terms found no hit:\t['cg24975666', 'cg24975723', 'cg24975769', 'cg24975821', 'cg24975834', 'cg24975842', 'cg24975898', '\n",
      "1000 input query terms found no hit:\t['cg25033470', 'cg25033515', 'cg25033719', 'cg25033759', 'cg25033767', 'cg25033817', 'cg25033972', '\n",
      "1000 input query terms found no hit:\t['cg25104234', 'cg25104254', 'cg25104266', 'cg25104397', 'cg25104430', 'cg25104437', 'cg25104457', '\n",
      "1000 input query terms found no hit:\t['cg25164996', 'cg25165049', 'cg25165060', 'cg25165144', 'cg25165171', 'cg25165199', 'cg25165358', '\n",
      "1000 input query terms found no hit:\t['cg25233208', 'cg25233299', 'cg25233302', 'cg25233308', 'cg25233339', 'cg25233520', 'cg25233534', '\n",
      "1000 input query terms found no hit:\t['cg25307318', 'cg25307332', 'cg25307371', 'cg25307468', 'cg25307471', 'cg25307556', 'cg25307593', '\n",
      "1000 input query terms found no hit:\t['cg25366139', 'cg25366140', 'cg25366141', 'cg25366155', 'cg25366157', 'cg25366315', 'cg25366390', '\n",
      "1000 input query terms found no hit:\t['cg25426302', 'cg25426350', 'cg25426560', 'cg25426563', 'cg25426630', 'cg25426716', 'cg25426722', '\n",
      "1000 input query terms found no hit:\t['cg25492620', 'cg25492645', 'cg25492727', 'cg25492803', 'cg25492875', 'cg25492971', 'cg25492985', '\n",
      "1000 input query terms found no hit:\t['cg25561581', 'cg25561639', 'cg25561754', 'cg25561762', 'cg25561792', 'cg25561821', 'cg25561875', '\n",
      "1000 input query terms found no hit:\t['cg25620760', 'cg25620797', 'cg25620901', 'cg25621037', 'cg25621094', 'cg25621096', 'cg25621182', '\n",
      "1000 input query terms found no hit:\t['cg25689940', 'cg25689955', 'cg25690016', 'cg25690202', 'cg25690209', 'cg25690265', 'cg25690537', '\n",
      "1000 input query terms found no hit:\t['cg25753693', 'cg25753720', 'cg25753722', 'cg25753817', 'cg25753841', 'cg25753915', 'cg25753990', '\n",
      "1000 input query terms found no hit:\t['cg25821436', 'cg25821437', 'cg25821781', 'cg25821783', 'cg25821785', 'cg25821794', 'cg25821896', '\n",
      "1000 input query terms found no hit:\t['cg25900614', 'cg25900631', 'cg25900813', 'cg25900902', 'cg25900943', 'cg25900995', 'cg25901000', '\n",
      "1000 input query terms found no hit:\t['cg25963980', 'cg25964007', 'cg25964032', 'cg25964035', 'cg25964040', 'cg25964067', 'cg25964180', '\n",
      "1000 input query terms found no hit:\t['cg26026748', 'cg26026821', 'cg26026951', 'cg26027052', 'cg26027170', 'cg26027205', 'cg26027233', '\n",
      "1000 input query terms found no hit:\t['cg26092180', 'cg26092192', 'cg26092233', 'cg26092253', 'cg26092367', 'cg26092468', 'cg26092564', '\n",
      "1000 input query terms found no hit:\t['cg26147279', 'cg26147311', 'cg26147335', 'cg26147338', 'cg26147351', 'cg26147372', 'cg26147459', '\n",
      "1000 input query terms found no hit:\t['cg26206748', 'cg26206894', 'cg26206910', 'cg26206948', 'cg26206973', 'cg26206990', 'cg26207017', '\n",
      "1000 input query terms found no hit:\t['cg26265933', 'cg26266041', 'cg26266046', 'cg26266098', 'cg26266099', 'cg26266274', 'cg26266308', '\n",
      "1000 input query terms found no hit:\t['cg26320173', 'cg26320219', 'cg26320244', 'cg26320372', 'cg26320410', 'cg26320504', 'cg26320601', '\n",
      "1000 input query terms found no hit:\t['cg26373812', 'cg26373874', 'cg26373922', 'cg26373942', 'cg26373949', 'cg26374058', 'cg26374101', '\n",
      "1000 input query terms found no hit:\t['cg26434225', 'cg26434278', 'cg26434290', 'cg26434328', 'cg26434370', 'cg26434376', 'cg26434400', '\n",
      "1000 input query terms found no hit:\t['cg26497185', 'cg26497264', 'cg26497342', 'cg26497348', 'cg26497354', 'cg26497399', 'cg26497532', '\n",
      "1000 input query terms found no hit:\t['cg26550872', 'cg26550873', 'cg26550936', 'cg26550938', 'cg26551026', 'cg26551072', 'cg26551092', '\n",
      "1000 input query terms found no hit:\t['cg26613140', 'cg26613239', 'cg26613312', 'cg26613361', 'cg26613367', 'cg26613586', 'cg26613648', '\n",
      "1000 input query terms found no hit:\t['cg26672794', 'cg26672807', 'cg26672952', 'cg26672959', 'cg26673012', 'cg26673053', 'cg26673070', '\n",
      "1000 input query terms found no hit:\t['cg26727693', 'cg26727768', 'cg26727807', 'cg26727816', 'cg26727980', 'cg26728141', 'cg26728282', '\n",
      "1000 input query terms found no hit:\t['cg26794015', 'cg26794028', 'cg26794156', 'cg26794221', 'cg26794227', 'cg26794346', 'cg26794364', '\n",
      "1000 input query terms found no hit:\t['cg26844459', 'cg26844633', 'cg26844732', 'cg26844804', 'cg26844909', 'cg26844954', 'cg26845036', '\n",
      "1000 input query terms found no hit:\t['cg26900616', 'cg26900686', 'cg26900724', 'cg26900750', 'cg26900781', 'cg26900875', 'cg26900934', '\n",
      "1000 input query terms found no hit:\t['cg26956494', 'cg26956535', 'cg26956754', 'cg26956760', 'cg26956799', 'cg26956874', 'cg26957012', '\n",
      "1000 input query terms found no hit:\t['cg27017357', 'cg27017382', 'cg27017414', 'cg27017482', 'cg27017562', 'cg27017580', 'cg27017630', '\n",
      "1000 input query terms found no hit:\t['cg27075492', 'cg27075518', 'cg27075654', 'cg27075724', 'cg27075786', 'cg27075800', 'cg27075898', '\n",
      "1000 input query terms found no hit:\t['cg27134659', 'cg27134730', 'cg27134732', 'cg27134767', 'cg27134812', 'cg27134816', 'cg27134827', '\n",
      "1000 input query terms found no hit:\t['cg27196131', 'cg27196158', 'cg27196194', 'cg27196273', 'cg27196467', 'cg27196496', 'cg27196659', '\n",
      "1000 input query terms found no hit:\t['cg27262398', 'cg27262412', 'cg27262415', 'cg27262488', 'cg27262702', 'cg27262717', 'cg27262735', '\n",
      "1000 input query terms found no hit:\t['cg27315497', 'cg27315555', 'cg27315556', 'cg27315579', 'cg27315601', 'cg27315633', 'cg27315635', '\n",
      "1000 input query terms found no hit:\t['cg27376924', 'cg27376941', 'cg27377213', 'cg27377289', 'cg27377339', 'cg27377352', 'cg27377450', '\n",
      "1000 input query terms found no hit:\t['cg27437721', 'cg27437868', 'cg27437944', 'cg27437966', 'cg27438040', 'cg27438128', 'cg27438152', '\n",
      "1000 input query terms found no hit:\t['cg27503095', 'cg27503117', 'cg27503275', 'cg27503360', 'cg27503398', 'cg27503469', 'cg27503499', '\n",
      "1000 input query terms found no hit:\t['cg27558693', 'cg27558697', 'cg27558820', 'cg27558864', 'cg27558876', 'cg27559021', 'cg27559097', '\n",
      "1000 input query terms found no hit:\t['cg27620052', 'cg27620077', 'cg27620135', 'cg27620173', 'cg27620174', 'cg27620176', 'cg27620230', '\n",
      "1000 input query terms found no hit:\t['ch.1.4505949F', 'ch.1.4511733R', 'ch.1.4512450F', 'ch.1.4544771F', 'ch.1.4597487F', 'ch.1.4603470R\n",
      "1000 input query terms found no hit:\t['ch.17.655115F', 'ch.17.65853408R', 'ch.17.66175159R', 'ch.17.67353124R', 'ch.17.69555766F', 'ch.17\n",
      "827 input query terms found no hit:\t['ch.5.3178288F', 'ch.5.31803364R', 'ch.5.3183300R', 'ch.5.3224933F', 'ch.5.3268483F', 'ch.5.3276032\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 42 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Hepatitis', the least common label is '0.0' with 21 occurrences. This represents 50.00% of the dataset.\n",
      "The distribution of the feature 'Hepatitis' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE113392</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "0  GSE113392       True          True      False    False       False   \n",
       "\n",
       "   sample_size note  \n",
       "0           42       "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.295070023Z",
     "start_time": "2024-01-10T21:40:21.338515691Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +: 'NoneType' and 'str'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [53]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m merged_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, \u001b[43mcondition_best_cohort\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m))\n\u001b[1;32m      2\u001b[0m merged_data\u001b[38;5;241m.\u001b[39mhead()\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for +: 'NoneType' and 'str'"
     ]
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.358779256Z",
     "start_time": "2024-01-10T21:40:27.271337819Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n",
    "\n",
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_batch_effect(X):\n",
    "    \"\"\"\n",
    "    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n",
    "\n",
    "    Args:\n",
    "    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n",
    "\n",
    "    Returns:\n",
    "    bool: True if a potential batch effect is detected, False otherwise.\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "\n",
    "    # Computing XX^T\n",
    "    XXt = np.dot(X, X.T)\n",
    "\n",
    "    # Compute the eigenvalues of XX^T\n",
    "    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n",
    "    eigen_values = sorted(eigen_values, reverse=True)\n",
    "\n",
    "    # Check for large gaps in the eigenvalues\n",
    "    for i in range(len(eigen_values) - 1):\n",
    "        gap = eigen_values[i] - eigen_values[i + 1]\n",
    "        if gap > 1 / n_samples:  # You may need to adjust this threshold\n",
    "            return True\n",
    "\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546590575Z",
     "start_time": "2024-01-10T21:40:27.298495045Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sparse_lmm import VariableSelection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546733889Z",
     "start_time": "2024-01-10T21:40:27.535249532Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Select appropriate models based on whether the dataset has batch effect.\n",
    "# We experiment on two models for each branch. We will decide which one to choose later.\n",
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n",
    "    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n",
    "    indices = np.arange(X.shape[0])\n",
    "    np.random.shuffle(indices)\n",
    "\n",
    "    fold_size = len(X) // k\n",
    "    performances = []\n",
    "\n",
    "    for i in range(k):\n",
    "        # Split data into train and test based on the current fold\n",
    "        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n",
    "        train_indices = np.setdiff1d(indices, test_indices)\n",
    "\n",
    "        X_train, X_test = X[train_indices], X[test_indices]\n",
    "        Y_train, Y_test = Y[train_indices], Y[test_indices]\n",
    "        Z_train, Z_test = Z[train_indices], Z[test_indices]\n",
    "\n",
    "        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n",
    "        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n",
    "\n",
    "        # model = model_constructor(**model_params)\n",
    "        model = ResidualizationRegressor(model_constructor, model_params)\n",
    "        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n",
    "        predictions = model.predict(normalized_X_test, normalized_Z_test)\n",
    "\n",
    "        if target_type == 'binary':\n",
    "            predictions = (predictions > 0.5).astype(int)\n",
    "            Y_test = (Y_test > 0.5).astype(int)\n",
    "            performance = accuracy_score(Y_test, predictions)\n",
    "        elif target_type == 'continuous':\n",
    "            performance = mean_squared_error(Y_test, predictions)\n",
    "\n",
    "        performances.append(performance)\n",
    "\n",
    "    cv_mean = np.mean(performances)\n",
    "    cv_std = np.std(performances)\n",
    "\n",
    "    if target_type == 'binary':\n",
    "        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n",
    "    else:\n",
    "        print(f'The cross-validation MSE is {(cv_mean * 100):.2f} ± {(cv_std * 100):.2f}')\n",
    "\n",
    "    return cv_mean, cv_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_data(X_train, X_test=None):\n",
    "    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n",
    "    and optionally the test data\"\"\"\n",
    "    mean = np.mean(X_train, axis=0)\n",
    "    std = np.std(X_train, axis=0)\n",
    "\n",
    "    # Handling columns with std = 0\n",
    "    std_no_zero = np.where(std == 0, 1, std)\n",
    "\n",
    "    # Normalize X_train\n",
    "    X_train_normalized = (X_train - mean) / std_no_zero\n",
    "    # Set normalized values to 0 where std was 0\n",
    "    X_train_normalized[:, std == 0] = 0\n",
    "\n",
    "    if X_test is not None:\n",
    "        X_test_normalized = (X_test - mean) / std_no_zero\n",
    "        X_test_normalized[:, std == 0] = 0\n",
    "    else:\n",
    "        X_test_normalized = None\n",
    "\n",
    "    return X_train_normalized, X_test_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ResidualizationRegressor:\n",
    "    def __init__(self, regression_model_constructor, params=None):\n",
    "        if params is None:\n",
    "            params = {}\n",
    "        self.regression_model = regression_model_constructor(**params)\n",
    "        self.beta_Z = None  # Coefficients for regression of Y on Z\n",
    "        self.beta_X = None  # Coefficients for regression of residual on X\n",
    "        self.neg_log_p_values = None  # Negative logarithm of p-values\n",
    "        self.p_values = None  # Actual p-values\n",
    "\n",
    "    def _reshape_data(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the data to ensure it's in the correct format (2D array).\n",
    "\n",
    "        :param data: The input data (can be 1D or 2D array).\n",
    "        :return: Reshaped 2D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 1:\n",
    "            return data.reshape(-1, 1)\n",
    "        return data\n",
    "\n",
    "    def _reshape_output(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the output data to ensure it's in the correct format (1D array).\n",
    "\n",
    "        :param data: The output data (can be 1D or 2D array).\n",
    "        :return: Reshaped 1D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 2 and data.shape[1] == 1:\n",
    "            return data.ravel()\n",
    "        return data\n",
    "    \n",
    "    def fit(self, X, Y, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Y = self._reshape_data(Y)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        # Step 1: Linear regression of Y on Z\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n",
    "        Y_hat = Z_ones @ self.beta_Z\n",
    "        e_Y = Y - Y_hat  # Residual of Y\n",
    "\n",
    "        # Step 2: Regress the residual on X using the included regression model\n",
    "        self.regression_model.fit(X, e_Y)\n",
    "\n",
    "        # Obtain coefficients from the regression model\n",
    "        if hasattr(self.regression_model, 'coef_'):\n",
    "            self.beta_X = self.regression_model.coef_\n",
    "        elif hasattr(self.regression_model, 'getBeta'):\n",
    "            beta_output = self.regression_model.getBeta()\n",
    "            self.beta_X = self._reshape_output(beta_output)\n",
    "\n",
    "        # Obtain negative logarithm of p-values, if available\n",
    "        if hasattr(self.regression_model, 'getNegLogP'):\n",
    "            neg_log_p_output = self.regression_model.getNegLogP()\n",
    "            if neg_log_p_output is not None:\n",
    "                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n",
    "                self.p_values = np.exp(-self.neg_log_p_values)\n",
    "                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n",
    "                p_values_Z = np.full(Z.shape[1], np.nan)\n",
    "                self.p_values = np.concatenate((p_values_Z, self.p_values))\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, mean_squared_error\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:30.570471636Z",
     "start_time": "2024-01-10T21:40:27.546295159Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:47.575878022Z",
     "start_time": "2024-01-10T21:40:30.570907992Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.663737008Z",
     "start_time": "2024-01-10T21:40:47.576021186Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n",
    "                     threshold: float = 0.05, save_output: bool = True,\n",
    "                     output_dir: str = './output', model_id: int = 1) -> None:\n",
    "    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n",
    "    consists of one variable about condition and multiple variables about genetic factors.\n",
    "    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n",
    "    p-values or non-zero coefficients, depending on the availability of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    model (Any): The trained regression Model.\n",
    "    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n",
    "    trait (str): The target trait of interest.\n",
    "    condition (str): The specific condition to examine within the model.\n",
    "    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n",
    "    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n",
    "    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n",
    "    model_id (int): The index of the model, 1 or 2.\n",
    "\n",
    "    Returns:\n",
    "    None: This function does not return anything but prints and optionally saves the output.\n",
    "    \"\"\"\n",
    "    coefficients = model.get_coefficients().reshape(-1).tolist()\n",
    "    p_values = model.get_p_values()\n",
    "    if p_values is None:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients\n",
    "        })\n",
    "    else:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients,\n",
    "            'p_value': p_values.reshape(-1).tolist()\n",
    "        })\n",
    "    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n",
    "\n",
    "    print(f\"Effect of the condition on the target variable:\")\n",
    "    print(f\"Variable: {condition}\")\n",
    "    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n",
    "    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n",
    "    if p_values is None:\n",
    "        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n",
    "        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n",
    "        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n",
    "            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n",
    "    else:\n",
    "        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n",
    "        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n",
    "        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n",
    "        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n",
    "        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n",
    "            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n",
    "\n",
    "    print(significant_genes_sorted.to_string(index=False))\n",
    "\n",
    "    # Optionally, save this to a CSV file\n",
    "    if save_output:\n",
    "        significant_genes_sorted.to_csv(\n",
    "            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.673565176Z",
     "start_time": "2024-01-10T21:40:51.665291317Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.730825913Z",
     "start_time": "2024-01-10T21:40:51.675049728Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
