{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "# Set your preferred name\n",
    "USER = \"Jiayi\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output'\n",
    "TRAIT = 'Kidney Chromophobe'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.928143012Z",
     "start_time": "2024-01-10T21:37:25.918685567Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.554597756Z",
     "start_time": "2024-01-10T21:37:26.530982646Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Kidney_Chromophobe_(KICH)-20240402T170820Z-001',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)',\n",
       " 'TCGA_Melanoma_(SKCM)-20240402T170844Z-001']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.999521704Z",
     "start_time": "2024-01-10T21:37:26.987471072Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Kidney_Chromophobe_(KICH)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.KICH.sampleMap_KICH_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.KICH.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.180581225Z",
     "start_time": "2024-01-10T21:37:27.217363685Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.185775284Z",
     "start_time": "2024-01-10T21:37:30.181745846Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'additional_pharmaceutical_therapy',\n",
       " 'additional_radiation_therapy',\n",
       " 'additional_surgery_locoregional_procedure',\n",
       " 'additional_surgery_metastatic_procedure',\n",
       " 'age_at_initial_pathologic_diagnosis']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.269854441Z",
     "start_time": "2024-01-10T21:37:30.184225525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_surgery_locoregional_procedure', 'additional_surgery_metastatic_procedure', 'age_at_initial_pathologic_diagnosis', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'clinical_M', 'days_to_additional_surgery_metastatic_procedure', 'days_to_birth', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_after_initial_treatment', 'eastern_cancer_oncology_group', 'followup_case_report_form_submission_reason', 'followup_treatment_success', 'form_completion_date', 'gender', 'hemoglobin_result', 'histological_type', 'history_of_neoadjuvant_treatment', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'intermediate_dimension', 'is_ffpe', 'karnofsky_performance_score', 'lactate_dehydrogenase_result', 'laterality', 'longest_dimension', 'lost_follow_up', 'lymph_node_examined_count', 'new_tumor_event_after_initial_treatment', 'number_of_lymphnodes_positive', 'number_pack_years_smoked', 'other_dx', 'pathologic_M', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'percent_tumor_sarcomatoid', 'performance_status_scale_timing', 'person_neoplasm_cancer_status', 'platelet_qualitative_result', 'presence_of_sarcomatoid_features', 'primary_lymph_node_presentation_assessment', 'primary_therapy_outcome_success', 'radiation_therapy', 'sample_type', 'sample_type_id', 'serum_calcium_result', 'shortest_dimension', 'stopped_smoking_year', 'system_version', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tobacco_smoking_history', 'tumor_tissue_site', 'vial_number', 'vital_status', 'white_cell_count_result', 'year_of_initial_pathologic_diagnosis', 'year_of_tobacco_smoking_onset', '_GENOMIC_ID_TCGA_KICH_PDMRNAseq', '_GENOMIC_ID_TCGA_KICH_exp_HiSeqV2_percentile', '_GENOMIC_ID_TCGA_KICH_gistic2thd', '_GENOMIC_ID_TCGA_KICH_mutation_bcgsc_gene', '_GENOMIC_ID_TCGA_KICH_exp_HiSeqV2', '_GENOMIC_ID_TCGA_KICH_RPPA', '_GENOMIC_ID_TCGA_KICH_miRNA_HiSeq', '_GENOMIC_ID_TCGA_KICH_mutation_bcm_gene', '_GENOMIC_ID_TCGA_KICH_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_KICH_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_KICH_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_KICH_hMethyl450', '_GENOMIC_ID_TCGA_KICH_mutation_broad_gene', '_GENOMIC_ID_data/public/TCGA/KICH/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_KICH_gistic2']\\n\""
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270036796Z",
     "start_time": "2024-01-10T21:37:30.227180075Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = [ 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = [ 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270232386Z",
     "start_time": "2024-01-10T21:37:30.227282564Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'age_at_initial_pathologic_diagnosis': [57, 67, 67, 56, 69],\n",
       " 'days_to_birth': [-20849, -24650, -24650, -20768, -25267],\n",
       " 'year_of_initial_pathologic_diagnosis': [2000, 2000, 2000, 2000, 2001]}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270339062Z",
     "start_time": "2024-01-10T21:37:30.227412883Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270520195Z",
     "start_time": "2024-01-10T21:37:30.227467704Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gender': ['FEMALE', 'FEMALE', 'FEMALE', 'FEMALE', 'MALE']}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270626721Z",
     "start_time": "2024-01-10T21:37:30.227530279Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270719762Z",
     "start_time": "2024-01-10T21:37:30.227585330Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.001977877Z",
     "start_time": "2024-01-10T21:37:30.227635342Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12 input query terms found dup hits:\t[('GTF2IP1', 2), ('RBMY1A3P', 3), ('RPL31P11', 2), ('HERC2P2', 3), ('WASH3P', 3), ('NUDT9P1', 2), ('\n",
      "154 input query terms found no hit:\t['C16orf13', 'C16orf11', 'LOC100272146', 'LOC339240', 'NACAP1', 'LOC441204', 'KLRA1', 'FAM183A', 'FA\n",
      "10 input query terms found dup hits:\t[('SUGT1P1', 2), ('PTPRVP', 2), ('SNORA62', 3), ('IFITM4P', 7), ('HLA-DRB6', 2), ('FUNDC2P2', 2), ('\n",
      "190 input query terms found no hit:\t['NARFL', 'NFKBIL2', 'LOC150197', 'TMEM84', 'LOC162632', 'PPPDE1', 'PPPDE2', 'C1orf38', 'C1orf31', '\n",
      "11 input query terms found dup hits:\t[('PIP5K1P1', 2), ('HBD', 2), ('PPP1R2P1', 9), ('HSD17B7P2', 2), ('RPSAP9', 2), ('SNORD68', 2), ('SN\n",
      "149 input query terms found no hit:\t['FAM153C', 'C9orf167', 'CLK2P', 'CCDC76', 'CCDC75', 'CCDC72', 'HIST3H2BB', 'PRAC', 'LOC285780', 'LO\n",
      "15 input query terms found dup hits:\t[('SNORD58C', 2), ('UOX', 2), ('UBE2Q2P1', 3), ('PPP4R1L', 2), ('SNORD63', 3), ('ESPNP', 2), ('HBBP1\n",
      "158 input query terms found no hit:\t['C16orf91', 'LOC100130581', 'SFRS2', 'C16orf5', 'C16orf7', 'C16orf3', 'C21orf7', 'C21orf2', 'LOC389\n",
      "15 input query terms found dup hits:\t[('FAM66D', 3), ('FAM66A', 2), ('THSD1P1', 2), ('EEF1DP3', 2), ('PGM5P2', 2), ('UBE2MP1', 2), ('HAR1\n",
      "169 input query terms found no hit:\t['LOC284551', 'LOC285548', 'LOC728410', 'LOC541473', 'DULLARD', 'KIAA0368', 'EFTUD1', 'TWISTNB', 'SF\n",
      "13 input query terms found dup hits:\t[('S100A7L2', 2), ('POM121L8P', 2), ('MEG8', 2), ('KIR3DX1', 5), ('RFPL1S', 2), ('SNORD91B', 2), ('C\n",
      "165 input query terms found no hit:\t['TMEM188', 'PDZD3', 'FAM102B', 'FAM102A', 'SMCR7L', 'G6PC', 'OSTCL', 'LOC653544', 'LOC653545', 'USP\n",
      "16 input query terms found dup hits:\t[('PCNAP1', 2), ('SNORA63', 6), ('SERHL', 2), ('CEACAM22P', 2), ('SNORA16A', 2), ('FAM41AY1', 2), ('\n",
      "147 input query terms found no hit:\t['LRRC37A4', 'LOC100131726', 'CPSF3L', 'COL4A3BP', 'PAR1', 'LOC92973', 'MICALCL', 'SMCR7', 'HIST4H4'\n",
      "15 input query terms found dup hits:\t[('MBL1P', 2), ('SDHAP3', 2), ('PSORS1C3', 8), ('MYADML', 2), ('POM121L10P', 2), ('HLA-J', 9), ('HLA\n",
      "153 input query terms found no hit:\t['LOC100131193', 'KIAA1109', 'DUS2L', 'MYST1', 'IGLL3', 'C12orf69', 'NEURL', 'FAM188B', 'FLJ32063', \n",
      "9 input query terms found dup hits:\t[('PLGLA', 2), ('PA2G4P4', 2), ('RPL23AP7', 2), ('RPL23AP64', 3), ('RPSAP52', 2), ('TPRXL', 2), ('DP\n",
      "177 input query terms found no hit:\t['LOC283314', 'RBM9', 'C20orf118', 'LOC150786', 'C20orf112', 'C20orf111', 'C20orf117', 'C20orf114', \n",
      "18 input query terms found dup hits:\t[('TPTE2P1', 2), ('FOLH1B', 2), ('ID2B', 2), ('MSX2P1', 2), ('RAET1K', 2), ('BMS1P4', 2), ('RPL23AP8\n",
      "175 input query terms found no hit:\t['MYEOV2', 'FAM35B2', 'C20orf85', 'FAM175A', 'C17orf73', 'C17orf72', 'C17orf71', 'C17orf70', 'C17orf\n",
      "19 input query terms found dup hits:\t[('CATSPER2P1', 2), ('AQP7P1', 2), ('ATP8B5P', 2), ('OR7E91P', 2), ('ZNF204P', 2), ('CECR7', 2), ('M\n",
      "144 input query terms found no hit:\t['BASE', 'C14orf21', 'C14orf23', 'FAM5B', 'TROVE2', 'C18orf62', 'LOC401052', 'GNB2L1', 'FAM19A4', 'F\n",
      "13 input query terms found dup hits:\t[('P2RX6P', 2), ('RP9P', 2), ('NRADDP', 2), ('MORF4', 2), ('SNORA75', 6), ('SNORA72', 2), ('FMO6P', \n",
      "168 input query terms found no hit:\t['TTC15', 'C4orf29', 'FAM48A', 'C4orf21', 'C3orf65', 'C4orf23', 'C3orf67', 'LOC144742', 'C3orf63', '\n",
      "19 input query terms found dup hits:\t[('RPL13AP17', 2), ('DNAJB3', 2), ('SNORA58', 2), ('PMCHL1', 2), ('FAM95B1', 2), ('TMEM191A', 2), ('\n",
      "144 input query terms found no hit:\t['LOC254312', 'ZNF542', 'CSDAP1', 'FAM73A', 'FAM73B', 'TCTE3', 'ZNF37B', 'LOC728875', 'KIAA1826', 'P\n",
      "15 input query terms found dup hits:\t[('ZFP91-CNTF', 2), ('PRSS30P', 3), ('GSTM2P1', 2), ('C3P1', 2), ('NCF1B', 2), ('SNORD114-4', 2), ('\n",
      "138 input query terms found no hit:\t['LOC100129716', 'C14orf184', 'C14orf181', 'C14orf183', 'C14orf182', 'C7orf26', 'C7orf27', 'C7orf23'\n",
      "14 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P3', 2), ('ANXA2P2', 2), ('ZNF286B', 2), ('SAA3P', 2), ('GEMIN8P4', 2), ('DI\n",
      "132 input query terms found no hit:\t['AKR1CL1', 'CCDC21', 'CCDC23', 'TMEM111', 'APOB48R', 'LOC595101', 'KIAA1430', 'KIAA1432', 'ORC4L', \n",
      "11 input query terms found dup hits:\t[('TCP10L2', 2), ('FTLP10', 2), ('TPTE2P3', 2), ('VENTXP1', 2), ('XKRY', 2), ('HTR7P1', 2), ('STAG3L\n",
      "175 input query terms found no hit:\t['DFNB59', 'LOC100170939', 'FAM36A', 'C2orf70', 'C2orf71', 'C2orf77', 'C2orf79', 'LOC100128191', 'HR\n",
      "16 input query terms found dup hits:\t[('ZNF702P', 2), ('BRD7P3', 2), ('SCARNA17', 2), ('GABARAPL3', 2), ('MT1IP', 2), ('ADAM6', 3), ('PDZ\n",
      "142 input query terms found no hit:\t['C9orf30', 'PEG3AS', 'C9orf37', 'TXNDC3', 'LOC442308', 'FAM46D', 'FAM46A', 'FAM46B', 'FAM46C', 'WDR\n",
      "10 input query terms found dup hits:\t[('FAM197Y2', 3), ('SNORA71B', 2), ('PRKY', 2), ('NAT8B', 3), ('VENTXP7', 2), ('SPRR2C', 2), ('RPL23\n",
      "148 input query terms found no hit:\t['FCGR1B', 'RG9MTD2', 'RARS', 'LOC727677', 'CCRN4L', 'SEPT14', 'HKR1', 'C1orf112', 'LOC100270710', '\n",
      "16 input query terms found dup hits:\t[('OR6W1P', 3), ('LYPLA2P1', 6), ('RPL23AP53', 2), ('TRPC2', 2), ('RBMY2FP', 2), ('BAGE2', 2), ('POM\n",
      "150 input query terms found no hit:\t['SIP1', 'FAM129C', 'GIF', 'TMCO7', 'CXorf61', 'CXorf64', 'ZNF192', 'SEPP1', 'KIRREL', 'MGC16121', '\n",
      "17 input query terms found dup hits:\t[('ISCA1P1', 2), ('PCDHB19P', 3), ('NAPSB', 2), ('WASH7P', 2), ('TTTY11', 2), ('HCG4', 9), ('OR5E1P'\n",
      "135 input query terms found no hit:\t['PPYR1', 'FAM82A2', 'FLJ37453', 'FAM82A1', 'KIAA0947', 'MOBKL2B', 'MOBKL2C', 'MOBKL2A', 'CMAH', 'GR\n",
      "5 input query terms found dup hits:\t[('GGNBP1', 2), ('WASH5P', 2), ('RPL29P2', 2), ('CDR1', 2), ('OR10J3', 3)]\n",
      "78 input query terms found no hit:\t['C14orf174', 'C14orf176', 'C14orf177', 'C14orf179', 'PTPLB', 'SGEF', 'LOC400752', 'GRLF1', 'SPANXE'\n"
     ]
    }
   ],
   "source": [
    "import mygene\n",
    "\n",
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.108889749Z",
     "start_time": "2024-01-10T21:38:18.021199718Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-KL-8323-01</th>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.295492</td>\n",
       "      <td>-3.487426</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.359428</td>\n",
       "      <td>0.850422</td>\n",
       "      <td>0.58019</td>\n",
       "      <td>-0.739994</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>0.397618</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>0.731583</td>\n",
       "      <td>-1.308333</td>\n",
       "      <td>-0.79709</td>\n",
       "      <td>-1.328077</td>\n",
       "      <td>0.933073</td>\n",
       "      <td>2.015005</td>\n",
       "      <td>0.025067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-KL-8324-01</th>\n",
       "      <td>1</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0.581408</td>\n",
       "      <td>0.368474</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.217628</td>\n",
       "      <td>0.626922</td>\n",
       "      <td>0.06679</td>\n",
       "      <td>-0.058894</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.193178</td>\n",
       "      <td>0.513683</td>\n",
       "      <td>-0.912933</td>\n",
       "      <td>-0.32549</td>\n",
       "      <td>-1.035777</td>\n",
       "      <td>0.221073</td>\n",
       "      <td>1.600605</td>\n",
       "      <td>2.456767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-KL-8324-11</th>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>1.119008</td>\n",
       "      <td>2.198374</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.341628</td>\n",
       "      <td>0.681022</td>\n",
       "      <td>0.46319</td>\n",
       "      <td>0.057806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.837786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.006622</td>\n",
       "      <td>5.587783</td>\n",
       "      <td>0.113667</td>\n",
       "      <td>-0.03259</td>\n",
       "      <td>0.051323</td>\n",
       "      <td>1.403473</td>\n",
       "      <td>0.786105</td>\n",
       "      <td>3.303867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-KL-8325-01</th>\n",
       "      <td>1</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>0.572008</td>\n",
       "      <td>-1.889926</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.683828</td>\n",
       "      <td>0.971922</td>\n",
       "      <td>1.55029</td>\n",
       "      <td>-0.150194</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>0.414383</td>\n",
       "      <td>-2.250633</td>\n",
       "      <td>1.13531</td>\n",
       "      <td>-1.257677</td>\n",
       "      <td>-1.076027</td>\n",
       "      <td>1.424005</td>\n",
       "      <td>-0.464633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-KL-8326-01</th>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>0.259208</td>\n",
       "      <td>-0.380726</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.992728</td>\n",
       "      <td>0.311022</td>\n",
       "      <td>0.65699</td>\n",
       "      <td>0.050806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.295086</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>1.982883</td>\n",
       "      <td>-0.845733</td>\n",
       "      <td>1.40521</td>\n",
       "      <td>-1.536277</td>\n",
       "      <td>0.100373</td>\n",
       "      <td>1.883005</td>\n",
       "      <td>0.013067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Kidney Chromophobe  Age  Gender  ARHGEF10L     HIF3A  \\\n",
       "sampleID                                                                \n",
       "TCGA-KL-8323-01                   1   57       0  -0.295492 -3.487426   \n",
       "TCGA-KL-8324-01                   1   67       0   0.581408  0.368474   \n",
       "TCGA-KL-8324-11                   0   67       0   1.119008  2.198374   \n",
       "TCGA-KL-8325-01                   1   56       0   0.572008 -1.889926   \n",
       "TCGA-KL-8326-01                   1   69       1   0.259208 -0.380726   \n",
       "\n",
       "                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n",
       "sampleID                                                          ...   \n",
       "TCGA-KL-8323-01 -0.531035  0.359428  0.850422  0.58019 -0.739994  ...   \n",
       "TCGA-KL-8324-01 -0.531035  1.217628  0.626922  0.06679 -0.058894  ...   \n",
       "TCGA-KL-8324-11 -0.531035  0.341628  0.681022  0.46319  0.057806  ...   \n",
       "TCGA-KL-8325-01 -0.531035  0.683828  0.971922  1.55029 -0.150194  ...   \n",
       "TCGA-KL-8326-01 -0.531035  0.992728  0.311022  0.65699  0.050806  ...   \n",
       "\n",
       "                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n",
       "sampleID                                                                     \n",
       "TCGA-KL-8323-01 -2.090786  0.397618 -0.748878  0.731583 -1.308333 -0.79709   \n",
       "TCGA-KL-8324-01 -2.090786 -0.086682 -0.193178  0.513683 -0.912933 -0.32549   \n",
       "TCGA-KL-8324-11 -0.837786 -0.086682  0.006622  5.587783  0.113667 -0.03259   \n",
       "TCGA-KL-8325-01 -2.090786 -0.086682 -0.748878  0.414383 -2.250633  1.13531   \n",
       "TCGA-KL-8326-01 -0.295086 -0.086682 -0.748878  1.982883 -0.845733  1.40521   \n",
       "\n",
       "                    TULP3     BCL6B     GSTK1      SELP  \n",
       "sampleID                                                 \n",
       "TCGA-KL-8323-01 -1.328077  0.933073  2.015005  0.025067  \n",
       "TCGA-KL-8324-01 -1.035777  0.221073  1.600605  2.456767  \n",
       "TCGA-KL-8324-11  0.051323  1.403473  0.786105  3.303867  \n",
       "TCGA-KL-8325-01 -1.257677 -1.076027  1.424005 -0.464633  \n",
       "TCGA-KL-8326-01 -1.536277  0.100373  1.883005  0.013067  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.112793384Z",
     "start_time": "2024-01-10T21:38:18.107415526Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 91 samples.\n",
      "For the feature 'Kidney Chromophobe', the least common label is '0' with 25 occurrences. This represents 27.47% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 42.5\n",
      "  50% (Median): 51.0\n",
      "  75%: 62.0\n",
      "Min: 17\n",
      "Max: 86\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0' with 39 occurrences. This represents 42.86% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.377898533Z",
     "start_time": "2024-01-10T21:38:18.111417452Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425336629Z",
     "start_time": "2024-01-10T21:38:31.419043732Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output\\Jiayi\\Kidney-Chromophobe\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE11024',\n",
       " 'GSE11151',\n",
       " 'GSE11447',\n",
       " 'GSE144082',\n",
       " 'GSE14670',\n",
       " 'GSE15641',\n",
       " 'GSE17746',\n",
       " 'GSE19949',\n",
       " 'GSE19982',\n",
       " 'GSE26574',\n",
       " 'GSE3',\n",
       " 'GSE40911',\n",
       " 'GSE40912',\n",
       " 'GSE40914',\n",
       " 'GSE4125',\n",
       " 'GSE42977',\n",
       " 'GSE57162',\n",
       " 'GSE6280',\n",
       " 'GSE68606',\n",
       " 'GSE8271',\n",
       " 'GSE95425']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Kidney-Chromophobe\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:02.935079426Z",
     "start_time": "2024-01-10T21:39:02.923698385Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene signatures of progression and metastasis in renal cell cancer\"\n",
      "!Series_summary\t\"In order to address the progression, metastasis, and clinical heterogeneity of renal cell cancer (RCC), transcriptional profiling with oligonucleotide microarrays (22,283 genes) was done on 49 RCC tumors, 20 non-RCC renal tumors, and 23 normal kidney samples. Samples were clustered based on gene expression profiles and specific gene sets for each renal tumor type were identified. Gene expression was correlated to disease progression and a metastasis gene signature was derived. Gene signatures were identified for each tumor type with 100% accuracy. Differentially expressed genes during early tumor formation and tumor progression to metastatic RCC were found. Subsets of these genes code for secreted proteins and membrane receptors and are both potential therapeutic or diagnostic targets. A gene pattern (\"\"metastatic signature\"\") derived from primary tumors was very accurate in classifying tumors with and without metastases at the time of surgery. A previously described \"\"global\"\" metastatic signature derived by another group from various non-RCC tumors was validated in RCC. Unlike previous studies, we describe highly accurate and externally validated gene signatures for RCC subtypes and other renal tumors. Interestingly, the gene expression of primary tumors provides us information about the metastatic status in the respective patients and has the potential, if prospectively validated, to enrich the armamentarium of diagnostic tests in RCC. We validated in RCC, for the first time, a previously described metastatic signature and further showed the feasibility of applying a gene signature across different microarray platforms. Transcriptional profiling allows a better appreciation of the molecular and clinical heterogeneity in RCC.\"\n",
      "!Series_overall_design\t\"We used the following tissue samples to obtain transcriptional profiling of kidney tumors using Affymetrix HGU-133A chips: 23 Normal, 32 clear cell RCC (cRCC), 11 papillary RCC (pRCC), 6 chromophobe RCC (chrRCC), 12 Oncocytoma (OC), and 8 transitional cell carcinoma (TCC).\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"The supplementary file 'GSE15641_mas5_data.txt' contains MAS5 signal values for the Samples included in Series GSE15641.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM391107</th>\n",
       "      <th>GSM391108</th>\n",
       "      <th>GSM391109</th>\n",
       "      <th>GSM391110</th>\n",
       "      <th>GSM391111</th>\n",
       "      <th>GSM391112</th>\n",
       "      <th>GSM391113</th>\n",
       "      <th>GSM391114</th>\n",
       "      <th>GSM391115</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM391189</th>\n",
       "      <th>GSM391190</th>\n",
       "      <th>GSM391191</th>\n",
       "      <th>GSM391192</th>\n",
       "      <th>GSM391193</th>\n",
       "      <th>GSM391194</th>\n",
       "      <th>GSM391195</th>\n",
       "      <th>GSM391196</th>\n",
       "      <th>GSM391197</th>\n",
       "      <th>GSM391198</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "      <td>tissue: Kidney</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession       GSM391107       GSM391108  \\\n",
       "0  !Sample_characteristics_ch1  tissue: Kidney  tissue: Kidney   \n",
       "\n",
       "        GSM391109       GSM391110       GSM391111       GSM391112  \\\n",
       "0  tissue: Kidney  tissue: Kidney  tissue: Kidney  tissue: Kidney   \n",
       "\n",
       "        GSM391113       GSM391114       GSM391115  ...       GSM391189  \\\n",
       "0  tissue: Kidney  tissue: Kidney  tissue: Kidney  ...  tissue: Kidney   \n",
       "\n",
       "        GSM391190       GSM391191       GSM391192       GSM391193  \\\n",
       "0  tissue: Kidney  tissue: Kidney  tissue: Kidney  tissue: Kidney   \n",
       "\n",
       "        GSM391194       GSM391195       GSM391196       GSM391197  \\\n",
       "0  tissue: Kidney  tissue: Kidney  tissue: Kidney  tissue: Kidney   \n",
       "\n",
       "        GSM391198  \n",
       "0  tissue: Kidney  \n",
       "\n",
       "[1 rows x 93 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE15641\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'tissue: Kidney'], dtype=object)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Kidney':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# 这个函数的目的是将年龄的字符串表示转换为一个连续的数值型表示。如果年龄未知（例如，标记为'n.a.'），则返回None。\n",
    "# 函数尝试从传入的字符串中提取出一个整数作为年龄值。如果字符串的格式不符合预期，导致提取失败，同样返回None。\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# 这个函数将性别的字符串表示转换为二进制值，其中“female”对应1，“male”对应0。如果性别未知或字符串不符合预期格式，则返回None。\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM391107</th>\n",
       "      <th>GSM391108</th>\n",
       "      <th>GSM391109</th>\n",
       "      <th>GSM391110</th>\n",
       "      <th>GSM391111</th>\n",
       "      <th>GSM391112</th>\n",
       "      <th>GSM391113</th>\n",
       "      <th>GSM391114</th>\n",
       "      <th>GSM391115</th>\n",
       "      <th>GSM391116</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM391189</th>\n",
       "      <th>GSM391190</th>\n",
       "      <th>GSM391191</th>\n",
       "      <th>GSM391192</th>\n",
       "      <th>GSM391193</th>\n",
       "      <th>GSM391194</th>\n",
       "      <th>GSM391195</th>\n",
       "      <th>GSM391196</th>\n",
       "      <th>GSM391197</th>\n",
       "      <th>GSM391198</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM391107  GSM391108  GSM391109  GSM391110  GSM391111  \\\n",
       "Kidney Chromophobe          1          1          1          1          1   \n",
       "\n",
       "                    GSM391112  GSM391113  GSM391114  GSM391115  GSM391116  \\\n",
       "Kidney Chromophobe          1          1          1          1          1   \n",
       "\n",
       "                    ...  GSM391189  GSM391190  GSM391191  GSM391192  \\\n",
       "Kidney Chromophobe  ...          1          1          1          1   \n",
       "\n",
       "                    GSM391193  GSM391194  GSM391195  GSM391196  GSM391197  \\\n",
       "Kidney Chromophobe          1          1          1          1          1   \n",
       "\n",
       "                    GSM391198  \n",
       "Kidney Chromophobe          1  \n",
       "\n",
       "[1 rows x 92 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM391107</th>\n",
       "      <th>GSM391108</th>\n",
       "      <th>GSM391109</th>\n",
       "      <th>GSM391110</th>\n",
       "      <th>GSM391111</th>\n",
       "      <th>GSM391112</th>\n",
       "      <th>GSM391113</th>\n",
       "      <th>GSM391114</th>\n",
       "      <th>GSM391115</th>\n",
       "      <th>GSM391116</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM391189</th>\n",
       "      <th>GSM391190</th>\n",
       "      <th>GSM391191</th>\n",
       "      <th>GSM391192</th>\n",
       "      <th>GSM391193</th>\n",
       "      <th>GSM391194</th>\n",
       "      <th>GSM391195</th>\n",
       "      <th>GSM391196</th>\n",
       "      <th>GSM391197</th>\n",
       "      <th>GSM391198</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>679.53</td>\n",
       "      <td>349.90</td>\n",
       "      <td>711.66</td>\n",
       "      <td>946.14</td>\n",
       "      <td>1027.44</td>\n",
       "      <td>726.10</td>\n",
       "      <td>820.66</td>\n",
       "      <td>861.99</td>\n",
       "      <td>796.80</td>\n",
       "      <td>811.09</td>\n",
       "      <td>...</td>\n",
       "      <td>909.19</td>\n",
       "      <td>761.81</td>\n",
       "      <td>1159.35</td>\n",
       "      <td>850.96</td>\n",
       "      <td>862.82</td>\n",
       "      <td>840.91</td>\n",
       "      <td>1069.96</td>\n",
       "      <td>951.86</td>\n",
       "      <td>1140.19</td>\n",
       "      <td>1422.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>190.75</td>\n",
       "      <td>179.89</td>\n",
       "      <td>192.90</td>\n",
       "      <td>182.62</td>\n",
       "      <td>168.39</td>\n",
       "      <td>150.40</td>\n",
       "      <td>148.28</td>\n",
       "      <td>145.87</td>\n",
       "      <td>136.88</td>\n",
       "      <td>151.87</td>\n",
       "      <td>...</td>\n",
       "      <td>204.78</td>\n",
       "      <td>158.85</td>\n",
       "      <td>175.79</td>\n",
       "      <td>157.64</td>\n",
       "      <td>151.13</td>\n",
       "      <td>150.52</td>\n",
       "      <td>143.85</td>\n",
       "      <td>186.49</td>\n",
       "      <td>158.97</td>\n",
       "      <td>150.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>154.05</td>\n",
       "      <td>163.18</td>\n",
       "      <td>158.30</td>\n",
       "      <td>155.36</td>\n",
       "      <td>153.08</td>\n",
       "      <td>140.50</td>\n",
       "      <td>133.20</td>\n",
       "      <td>121.55</td>\n",
       "      <td>130.79</td>\n",
       "      <td>147.74</td>\n",
       "      <td>...</td>\n",
       "      <td>145.93</td>\n",
       "      <td>137.33</td>\n",
       "      <td>161.83</td>\n",
       "      <td>139.11</td>\n",
       "      <td>137.51</td>\n",
       "      <td>131.97</td>\n",
       "      <td>133.61</td>\n",
       "      <td>152.76</td>\n",
       "      <td>149.31</td>\n",
       "      <td>131.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>1851.76</td>\n",
       "      <td>1347.33</td>\n",
       "      <td>1278.88</td>\n",
       "      <td>1530.74</td>\n",
       "      <td>1695.97</td>\n",
       "      <td>1427.47</td>\n",
       "      <td>1573.13</td>\n",
       "      <td>1658.89</td>\n",
       "      <td>1423.71</td>\n",
       "      <td>1585.53</td>\n",
       "      <td>...</td>\n",
       "      <td>1125.78</td>\n",
       "      <td>1358.90</td>\n",
       "      <td>1091.68</td>\n",
       "      <td>1310.38</td>\n",
       "      <td>1997.77</td>\n",
       "      <td>1711.60</td>\n",
       "      <td>2078.28</td>\n",
       "      <td>1865.49</td>\n",
       "      <td>789.54</td>\n",
       "      <td>916.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>51.55</td>\n",
       "      <td>56.91</td>\n",
       "      <td>41.02</td>\n",
       "      <td>28.71</td>\n",
       "      <td>33.50</td>\n",
       "      <td>21.63</td>\n",
       "      <td>33.36</td>\n",
       "      <td>32.96</td>\n",
       "      <td>27.96</td>\n",
       "      <td>34.43</td>\n",
       "      <td>...</td>\n",
       "      <td>33.25</td>\n",
       "      <td>40.68</td>\n",
       "      <td>51.91</td>\n",
       "      <td>38.29</td>\n",
       "      <td>32.97</td>\n",
       "      <td>34.79</td>\n",
       "      <td>33.54</td>\n",
       "      <td>33.61</td>\n",
       "      <td>42.25</td>\n",
       "      <td>33.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>54.49</td>\n",
       "      <td>65.92</td>\n",
       "      <td>38.78</td>\n",
       "      <td>38.31</td>\n",
       "      <td>39.09</td>\n",
       "      <td>43.80</td>\n",
       "      <td>45.69</td>\n",
       "      <td>50.13</td>\n",
       "      <td>45.67</td>\n",
       "      <td>45.27</td>\n",
       "      <td>...</td>\n",
       "      <td>187.14</td>\n",
       "      <td>153.28</td>\n",
       "      <td>188.94</td>\n",
       "      <td>130.51</td>\n",
       "      <td>127.00</td>\n",
       "      <td>139.74</td>\n",
       "      <td>115.42</td>\n",
       "      <td>139.34</td>\n",
       "      <td>214.32</td>\n",
       "      <td>144.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>22.43</td>\n",
       "      <td>25.05</td>\n",
       "      <td>17.14</td>\n",
       "      <td>13.20</td>\n",
       "      <td>16.62</td>\n",
       "      <td>20.58</td>\n",
       "      <td>20.76</td>\n",
       "      <td>18.94</td>\n",
       "      <td>17.82</td>\n",
       "      <td>21.95</td>\n",
       "      <td>...</td>\n",
       "      <td>20.92</td>\n",
       "      <td>20.26</td>\n",
       "      <td>20.38</td>\n",
       "      <td>20.16</td>\n",
       "      <td>15.27</td>\n",
       "      <td>13.72</td>\n",
       "      <td>17.47</td>\n",
       "      <td>19.24</td>\n",
       "      <td>23.35</td>\n",
       "      <td>21.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>9.75</td>\n",
       "      <td>9.44</td>\n",
       "      <td>13.83</td>\n",
       "      <td>10.61</td>\n",
       "      <td>12.32</td>\n",
       "      <td>9.15</td>\n",
       "      <td>5.24</td>\n",
       "      <td>8.10</td>\n",
       "      <td>7.98</td>\n",
       "      <td>8.63</td>\n",
       "      <td>...</td>\n",
       "      <td>14.93</td>\n",
       "      <td>16.45</td>\n",
       "      <td>8.56</td>\n",
       "      <td>8.26</td>\n",
       "      <td>12.06</td>\n",
       "      <td>11.87</td>\n",
       "      <td>12.51</td>\n",
       "      <td>12.21</td>\n",
       "      <td>15.17</td>\n",
       "      <td>17.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>56.94</td>\n",
       "      <td>50.68</td>\n",
       "      <td>54.77</td>\n",
       "      <td>48.93</td>\n",
       "      <td>47.31</td>\n",
       "      <td>46.67</td>\n",
       "      <td>51.77</td>\n",
       "      <td>55.94</td>\n",
       "      <td>48.45</td>\n",
       "      <td>50.82</td>\n",
       "      <td>...</td>\n",
       "      <td>46.18</td>\n",
       "      <td>48.27</td>\n",
       "      <td>46.73</td>\n",
       "      <td>41.90</td>\n",
       "      <td>38.26</td>\n",
       "      <td>46.66</td>\n",
       "      <td>40.75</td>\n",
       "      <td>35.47</td>\n",
       "      <td>35.27</td>\n",
       "      <td>40.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>23.91</td>\n",
       "      <td>30.46</td>\n",
       "      <td>24.85</td>\n",
       "      <td>21.20</td>\n",
       "      <td>21.69</td>\n",
       "      <td>18.32</td>\n",
       "      <td>25.05</td>\n",
       "      <td>17.62</td>\n",
       "      <td>21.06</td>\n",
       "      <td>22.18</td>\n",
       "      <td>...</td>\n",
       "      <td>25.35</td>\n",
       "      <td>32.25</td>\n",
       "      <td>35.29</td>\n",
       "      <td>27.87</td>\n",
       "      <td>28.70</td>\n",
       "      <td>26.35</td>\n",
       "      <td>23.38</td>\n",
       "      <td>30.56</td>\n",
       "      <td>26.54</td>\n",
       "      <td>26.36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22283 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM391107  GSM391108  GSM391109  GSM391110  GSM391111  \\\n",
       "ID                                                                       \n",
       "1007_s_at           679.53     349.90     711.66     946.14    1027.44   \n",
       "1053_at             190.75     179.89     192.90     182.62     168.39   \n",
       "117_at              154.05     163.18     158.30     155.36     153.08   \n",
       "121_at             1851.76    1347.33    1278.88    1530.74    1695.97   \n",
       "1255_g_at            51.55      56.91      41.02      28.71      33.50   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "AFFX-ThrX-5_at       54.49      65.92      38.78      38.31      39.09   \n",
       "AFFX-ThrX-M_at       22.43      25.05      17.14      13.20      16.62   \n",
       "AFFX-TrpnX-3_at       9.75       9.44      13.83      10.61      12.32   \n",
       "AFFX-TrpnX-5_at      56.94      50.68      54.77      48.93      47.31   \n",
       "AFFX-TrpnX-M_at      23.91      30.46      24.85      21.20      21.69   \n",
       "\n",
       "                 GSM391112  GSM391113  GSM391114  GSM391115  GSM391116  ...  \\\n",
       "ID                                                                      ...   \n",
       "1007_s_at           726.10     820.66     861.99     796.80     811.09  ...   \n",
       "1053_at             150.40     148.28     145.87     136.88     151.87  ...   \n",
       "117_at              140.50     133.20     121.55     130.79     147.74  ...   \n",
       "121_at             1427.47    1573.13    1658.89    1423.71    1585.53  ...   \n",
       "1255_g_at            21.63      33.36      32.96      27.96      34.43  ...   \n",
       "...                    ...        ...        ...        ...        ...  ...   \n",
       "AFFX-ThrX-5_at       43.80      45.69      50.13      45.67      45.27  ...   \n",
       "AFFX-ThrX-M_at       20.58      20.76      18.94      17.82      21.95  ...   \n",
       "AFFX-TrpnX-3_at       9.15       5.24       8.10       7.98       8.63  ...   \n",
       "AFFX-TrpnX-5_at      46.67      51.77      55.94      48.45      50.82  ...   \n",
       "AFFX-TrpnX-M_at      18.32      25.05      17.62      21.06      22.18  ...   \n",
       "\n",
       "                 GSM391189  GSM391190  GSM391191  GSM391192  GSM391193  \\\n",
       "ID                                                                       \n",
       "1007_s_at           909.19     761.81    1159.35     850.96     862.82   \n",
       "1053_at             204.78     158.85     175.79     157.64     151.13   \n",
       "117_at              145.93     137.33     161.83     139.11     137.51   \n",
       "121_at             1125.78    1358.90    1091.68    1310.38    1997.77   \n",
       "1255_g_at            33.25      40.68      51.91      38.29      32.97   \n",
       "...                    ...        ...        ...        ...        ...   \n",
       "AFFX-ThrX-5_at      187.14     153.28     188.94     130.51     127.00   \n",
       "AFFX-ThrX-M_at       20.92      20.26      20.38      20.16      15.27   \n",
       "AFFX-TrpnX-3_at      14.93      16.45       8.56       8.26      12.06   \n",
       "AFFX-TrpnX-5_at      46.18      48.27      46.73      41.90      38.26   \n",
       "AFFX-TrpnX-M_at      25.35      32.25      35.29      27.87      28.70   \n",
       "\n",
       "                 GSM391194  GSM391195  GSM391196  GSM391197  GSM391198  \n",
       "ID                                                                      \n",
       "1007_s_at           840.91    1069.96     951.86    1140.19    1422.83  \n",
       "1053_at             150.52     143.85     186.49     158.97     150.36  \n",
       "117_at              131.97     133.61     152.76     149.31     131.13  \n",
       "121_at             1711.60    2078.28    1865.49     789.54     916.89  \n",
       "1255_g_at            34.79      33.54      33.61      42.25      33.67  \n",
       "...                    ...        ...        ...        ...        ...  \n",
       "AFFX-ThrX-5_at      139.74     115.42     139.34     214.32     144.41  \n",
       "AFFX-ThrX-M_at       13.72      17.47      19.24      23.35      21.44  \n",
       "AFFX-TrpnX-3_at      11.87      12.51      12.21      15.17      17.54  \n",
       "AFFX-TrpnX-5_at      46.66      40.75      35.47      35.27      40.27  \n",
       "AFFX-TrpnX-M_at      26.35      23.38      30.56      26.54      26.36  \n",
       "\n",
       "[22283 rows x 92 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1598_g_at', '160020_at', '1729_at', '1773_at', '177_at', '179_at', '1861_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM391107</th>\n",
       "      <th>GSM391108</th>\n",
       "      <th>GSM391109</th>\n",
       "      <th>GSM391110</th>\n",
       "      <th>GSM391111</th>\n",
       "      <th>GSM391112</th>\n",
       "      <th>GSM391113</th>\n",
       "      <th>GSM391114</th>\n",
       "      <th>GSM391115</th>\n",
       "      <th>GSM391116</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM391189</th>\n",
       "      <th>GSM391190</th>\n",
       "      <th>GSM391191</th>\n",
       "      <th>GSM391192</th>\n",
       "      <th>GSM391193</th>\n",
       "      <th>GSM391194</th>\n",
       "      <th>GSM391195</th>\n",
       "      <th>GSM391196</th>\n",
       "      <th>GSM391197</th>\n",
       "      <th>GSM391198</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>216.850</td>\n",
       "      <td>273.32</td>\n",
       "      <td>286.000000</td>\n",
       "      <td>270.790000</td>\n",
       "      <td>234.910000</td>\n",
       "      <td>354.430000</td>\n",
       "      <td>301.360000</td>\n",
       "      <td>291.760000</td>\n",
       "      <td>259.660000</td>\n",
       "      <td>373.820</td>\n",
       "      <td>...</td>\n",
       "      <td>221.61</td>\n",
       "      <td>210.39</td>\n",
       "      <td>238.120000</td>\n",
       "      <td>214.250000</td>\n",
       "      <td>342.850000</td>\n",
       "      <td>184.160000</td>\n",
       "      <td>186.550000</td>\n",
       "      <td>204.180000</td>\n",
       "      <td>208.02</td>\n",
       "      <td>181.880000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>1234.580</td>\n",
       "      <td>1600.46</td>\n",
       "      <td>1138.550000</td>\n",
       "      <td>1222.280000</td>\n",
       "      <td>1633.940000</td>\n",
       "      <td>979.330000</td>\n",
       "      <td>1415.140000</td>\n",
       "      <td>1156.080000</td>\n",
       "      <td>1633.880000</td>\n",
       "      <td>1346.710</td>\n",
       "      <td>...</td>\n",
       "      <td>978.94</td>\n",
       "      <td>716.96</td>\n",
       "      <td>643.230000</td>\n",
       "      <td>1527.600000</td>\n",
       "      <td>2448.750000</td>\n",
       "      <td>2136.700000</td>\n",
       "      <td>1987.630000</td>\n",
       "      <td>1182.570000</td>\n",
       "      <td>665.23</td>\n",
       "      <td>954.210000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>156.470</td>\n",
       "      <td>147.93</td>\n",
       "      <td>250.030000</td>\n",
       "      <td>209.760000</td>\n",
       "      <td>244.660000</td>\n",
       "      <td>224.050000</td>\n",
       "      <td>182.410000</td>\n",
       "      <td>220.980000</td>\n",
       "      <td>240.770000</td>\n",
       "      <td>214.970</td>\n",
       "      <td>...</td>\n",
       "      <td>100.28</td>\n",
       "      <td>102.09</td>\n",
       "      <td>101.090000</td>\n",
       "      <td>77.750000</td>\n",
       "      <td>92.030000</td>\n",
       "      <td>95.890000</td>\n",
       "      <td>95.840000</td>\n",
       "      <td>98.780000</td>\n",
       "      <td>92.76</td>\n",
       "      <td>96.770000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>114.730</td>\n",
       "      <td>96.43</td>\n",
       "      <td>115.990000</td>\n",
       "      <td>82.230000</td>\n",
       "      <td>91.990000</td>\n",
       "      <td>93.310000</td>\n",
       "      <td>95.910000</td>\n",
       "      <td>94.970000</td>\n",
       "      <td>94.620000</td>\n",
       "      <td>84.930</td>\n",
       "      <td>...</td>\n",
       "      <td>200.55</td>\n",
       "      <td>182.43</td>\n",
       "      <td>178.820000</td>\n",
       "      <td>178.160000</td>\n",
       "      <td>189.040000</td>\n",
       "      <td>149.580000</td>\n",
       "      <td>151.390000</td>\n",
       "      <td>149.030000</td>\n",
       "      <td>183.64</td>\n",
       "      <td>166.220000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>205.970</td>\n",
       "      <td>161.96</td>\n",
       "      <td>186.940000</td>\n",
       "      <td>175.210000</td>\n",
       "      <td>169.830000</td>\n",
       "      <td>180.700000</td>\n",
       "      <td>161.470000</td>\n",
       "      <td>168.280000</td>\n",
       "      <td>158.630000</td>\n",
       "      <td>178.560</td>\n",
       "      <td>...</td>\n",
       "      <td>215.45</td>\n",
       "      <td>179.03</td>\n",
       "      <td>193.500000</td>\n",
       "      <td>221.900000</td>\n",
       "      <td>189.570000</td>\n",
       "      <td>204.720000</td>\n",
       "      <td>188.040000</td>\n",
       "      <td>216.820000</td>\n",
       "      <td>217.65</td>\n",
       "      <td>204.090000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>76.150</td>\n",
       "      <td>73.69</td>\n",
       "      <td>67.460000</td>\n",
       "      <td>59.640000</td>\n",
       "      <td>69.450000</td>\n",
       "      <td>55.650000</td>\n",
       "      <td>57.320000</td>\n",
       "      <td>52.890000</td>\n",
       "      <td>54.360000</td>\n",
       "      <td>68.740</td>\n",
       "      <td>...</td>\n",
       "      <td>68.87</td>\n",
       "      <td>77.84</td>\n",
       "      <td>71.570000</td>\n",
       "      <td>65.840000</td>\n",
       "      <td>68.270000</td>\n",
       "      <td>70.480000</td>\n",
       "      <td>56.580000</td>\n",
       "      <td>57.000000</td>\n",
       "      <td>77.04</td>\n",
       "      <td>76.120000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>229.670</td>\n",
       "      <td>204.86</td>\n",
       "      <td>171.430000</td>\n",
       "      <td>198.630000</td>\n",
       "      <td>208.860000</td>\n",
       "      <td>206.790000</td>\n",
       "      <td>214.290000</td>\n",
       "      <td>212.040000</td>\n",
       "      <td>268.640000</td>\n",
       "      <td>194.250</td>\n",
       "      <td>...</td>\n",
       "      <td>187.55</td>\n",
       "      <td>154.18</td>\n",
       "      <td>174.780000</td>\n",
       "      <td>206.150000</td>\n",
       "      <td>210.740000</td>\n",
       "      <td>154.120000</td>\n",
       "      <td>160.570000</td>\n",
       "      <td>159.500000</td>\n",
       "      <td>196.24</td>\n",
       "      <td>145.630000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>364.805</td>\n",
       "      <td>319.76</td>\n",
       "      <td>403.445000</td>\n",
       "      <td>459.975000</td>\n",
       "      <td>665.935000</td>\n",
       "      <td>317.120000</td>\n",
       "      <td>315.620000</td>\n",
       "      <td>302.040000</td>\n",
       "      <td>295.910000</td>\n",
       "      <td>407.785</td>\n",
       "      <td>...</td>\n",
       "      <td>418.80</td>\n",
       "      <td>336.44</td>\n",
       "      <td>332.555000</td>\n",
       "      <td>385.960000</td>\n",
       "      <td>415.850000</td>\n",
       "      <td>466.205000</td>\n",
       "      <td>484.980000</td>\n",
       "      <td>406.645000</td>\n",
       "      <td>327.05</td>\n",
       "      <td>349.415000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>110.290</td>\n",
       "      <td>115.06</td>\n",
       "      <td>104.916667</td>\n",
       "      <td>96.573333</td>\n",
       "      <td>95.296667</td>\n",
       "      <td>98.516667</td>\n",
       "      <td>102.956667</td>\n",
       "      <td>93.953333</td>\n",
       "      <td>96.336667</td>\n",
       "      <td>104.350</td>\n",
       "      <td>...</td>\n",
       "      <td>131.18</td>\n",
       "      <td>125.87</td>\n",
       "      <td>119.086667</td>\n",
       "      <td>134.856667</td>\n",
       "      <td>119.343333</td>\n",
       "      <td>130.183333</td>\n",
       "      <td>124.223333</td>\n",
       "      <td>123.346667</td>\n",
       "      <td>117.47</td>\n",
       "      <td>123.816667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>120.570</td>\n",
       "      <td>168.53</td>\n",
       "      <td>115.280000</td>\n",
       "      <td>149.220000</td>\n",
       "      <td>149.950000</td>\n",
       "      <td>170.530000</td>\n",
       "      <td>152.940000</td>\n",
       "      <td>166.620000</td>\n",
       "      <td>147.310000</td>\n",
       "      <td>135.680</td>\n",
       "      <td>...</td>\n",
       "      <td>180.89</td>\n",
       "      <td>138.73</td>\n",
       "      <td>134.330000</td>\n",
       "      <td>171.260000</td>\n",
       "      <td>160.680000</td>\n",
       "      <td>169.520000</td>\n",
       "      <td>178.070000</td>\n",
       "      <td>125.930000</td>\n",
       "      <td>321.24</td>\n",
       "      <td>187.640000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11727 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM391107  GSM391108    GSM391109    GSM391110    GSM391111  \\\n",
       "A1CF      216.850     273.32   286.000000   270.790000   234.910000   \n",
       "A2M      1234.580    1600.46  1138.550000  1222.280000  1633.940000   \n",
       "A4GALT    156.470     147.93   250.030000   209.760000   244.660000   \n",
       "A4GNT     114.730      96.43   115.990000    82.230000    91.990000   \n",
       "AAAS      205.970     161.96   186.940000   175.210000   169.830000   \n",
       "...           ...        ...          ...          ...          ...   \n",
       "ZXDB       76.150      73.69    67.460000    59.640000    69.450000   \n",
       "ZXDC      229.670     204.86   171.430000   198.630000   208.860000   \n",
       "ZYX       364.805     319.76   403.445000   459.975000   665.935000   \n",
       "ZZEF1     110.290     115.06   104.916667    96.573333    95.296667   \n",
       "ZZZ3      120.570     168.53   115.280000   149.220000   149.950000   \n",
       "\n",
       "         GSM391112    GSM391113    GSM391114    GSM391115  GSM391116  ...  \\\n",
       "A1CF    354.430000   301.360000   291.760000   259.660000    373.820  ...   \n",
       "A2M     979.330000  1415.140000  1156.080000  1633.880000   1346.710  ...   \n",
       "A4GALT  224.050000   182.410000   220.980000   240.770000    214.970  ...   \n",
       "A4GNT    93.310000    95.910000    94.970000    94.620000     84.930  ...   \n",
       "AAAS    180.700000   161.470000   168.280000   158.630000    178.560  ...   \n",
       "...            ...          ...          ...          ...        ...  ...   \n",
       "ZXDB     55.650000    57.320000    52.890000    54.360000     68.740  ...   \n",
       "ZXDC    206.790000   214.290000   212.040000   268.640000    194.250  ...   \n",
       "ZYX     317.120000   315.620000   302.040000   295.910000    407.785  ...   \n",
       "ZZEF1    98.516667   102.956667    93.953333    96.336667    104.350  ...   \n",
       "ZZZ3    170.530000   152.940000   166.620000   147.310000    135.680  ...   \n",
       "\n",
       "        GSM391189  GSM391190   GSM391191    GSM391192    GSM391193  \\\n",
       "A1CF       221.61     210.39  238.120000   214.250000   342.850000   \n",
       "A2M        978.94     716.96  643.230000  1527.600000  2448.750000   \n",
       "A4GALT     100.28     102.09  101.090000    77.750000    92.030000   \n",
       "A4GNT      200.55     182.43  178.820000   178.160000   189.040000   \n",
       "AAAS       215.45     179.03  193.500000   221.900000   189.570000   \n",
       "...           ...        ...         ...          ...          ...   \n",
       "ZXDB        68.87      77.84   71.570000    65.840000    68.270000   \n",
       "ZXDC       187.55     154.18  174.780000   206.150000   210.740000   \n",
       "ZYX        418.80     336.44  332.555000   385.960000   415.850000   \n",
       "ZZEF1      131.18     125.87  119.086667   134.856667   119.343333   \n",
       "ZZZ3       180.89     138.73  134.330000   171.260000   160.680000   \n",
       "\n",
       "          GSM391194    GSM391195    GSM391196  GSM391197   GSM391198  \n",
       "A1CF     184.160000   186.550000   204.180000     208.02  181.880000  \n",
       "A2M     2136.700000  1987.630000  1182.570000     665.23  954.210000  \n",
       "A4GALT    95.890000    95.840000    98.780000      92.76   96.770000  \n",
       "A4GNT    149.580000   151.390000   149.030000     183.64  166.220000  \n",
       "AAAS     204.720000   188.040000   216.820000     217.65  204.090000  \n",
       "...             ...          ...          ...        ...         ...  \n",
       "ZXDB      70.480000    56.580000    57.000000      77.04   76.120000  \n",
       "ZXDC     154.120000   160.570000   159.500000     196.24  145.630000  \n",
       "ZYX      466.205000   484.980000   406.645000     327.05  349.415000  \n",
       "ZZEF1    130.183333   124.223333   123.346667     117.47  123.816667  \n",
       "ZZZ3     169.520000   178.070000   125.930000     321.24  187.640000  \n",
       "\n",
       "[11727 rows x 92 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZSWIM8</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM391107</th>\n",
       "      <td>1.0</td>\n",
       "      <td>216.85</td>\n",
       "      <td>1234.58</td>\n",
       "      <td>156.47</td>\n",
       "      <td>114.73</td>\n",
       "      <td>205.97</td>\n",
       "      <td>275.87</td>\n",
       "      <td>89.78</td>\n",
       "      <td>166.720</td>\n",
       "      <td>181.854</td>\n",
       "      <td>...</td>\n",
       "      <td>139.83</td>\n",
       "      <td>136.890</td>\n",
       "      <td>164.60</td>\n",
       "      <td>88.83</td>\n",
       "      <td>124.41</td>\n",
       "      <td>76.15</td>\n",
       "      <td>229.67</td>\n",
       "      <td>364.805</td>\n",
       "      <td>110.290000</td>\n",
       "      <td>120.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391108</th>\n",
       "      <td>1.0</td>\n",
       "      <td>273.32</td>\n",
       "      <td>1600.46</td>\n",
       "      <td>147.93</td>\n",
       "      <td>96.43</td>\n",
       "      <td>161.96</td>\n",
       "      <td>237.88</td>\n",
       "      <td>74.43</td>\n",
       "      <td>202.045</td>\n",
       "      <td>166.652</td>\n",
       "      <td>...</td>\n",
       "      <td>137.27</td>\n",
       "      <td>127.165</td>\n",
       "      <td>131.09</td>\n",
       "      <td>77.84</td>\n",
       "      <td>164.05</td>\n",
       "      <td>73.69</td>\n",
       "      <td>204.86</td>\n",
       "      <td>319.760</td>\n",
       "      <td>115.060000</td>\n",
       "      <td>168.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391109</th>\n",
       "      <td>1.0</td>\n",
       "      <td>286.00</td>\n",
       "      <td>1138.55</td>\n",
       "      <td>250.03</td>\n",
       "      <td>115.99</td>\n",
       "      <td>186.94</td>\n",
       "      <td>243.51</td>\n",
       "      <td>72.76</td>\n",
       "      <td>204.975</td>\n",
       "      <td>149.652</td>\n",
       "      <td>...</td>\n",
       "      <td>129.74</td>\n",
       "      <td>175.435</td>\n",
       "      <td>142.49</td>\n",
       "      <td>50.41</td>\n",
       "      <td>112.01</td>\n",
       "      <td>67.46</td>\n",
       "      <td>171.43</td>\n",
       "      <td>403.445</td>\n",
       "      <td>104.916667</td>\n",
       "      <td>115.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391110</th>\n",
       "      <td>1.0</td>\n",
       "      <td>270.79</td>\n",
       "      <td>1222.28</td>\n",
       "      <td>209.76</td>\n",
       "      <td>82.23</td>\n",
       "      <td>175.21</td>\n",
       "      <td>187.69</td>\n",
       "      <td>72.96</td>\n",
       "      <td>161.780</td>\n",
       "      <td>151.998</td>\n",
       "      <td>...</td>\n",
       "      <td>108.38</td>\n",
       "      <td>171.055</td>\n",
       "      <td>166.73</td>\n",
       "      <td>54.93</td>\n",
       "      <td>80.23</td>\n",
       "      <td>59.64</td>\n",
       "      <td>198.63</td>\n",
       "      <td>459.975</td>\n",
       "      <td>96.573333</td>\n",
       "      <td>149.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391111</th>\n",
       "      <td>1.0</td>\n",
       "      <td>234.91</td>\n",
       "      <td>1633.94</td>\n",
       "      <td>244.66</td>\n",
       "      <td>91.99</td>\n",
       "      <td>169.83</td>\n",
       "      <td>216.02</td>\n",
       "      <td>68.80</td>\n",
       "      <td>172.450</td>\n",
       "      <td>129.822</td>\n",
       "      <td>...</td>\n",
       "      <td>95.99</td>\n",
       "      <td>150.990</td>\n",
       "      <td>159.86</td>\n",
       "      <td>76.54</td>\n",
       "      <td>100.93</td>\n",
       "      <td>69.45</td>\n",
       "      <td>208.86</td>\n",
       "      <td>665.935</td>\n",
       "      <td>95.296667</td>\n",
       "      <td>149.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391194</th>\n",
       "      <td>1.0</td>\n",
       "      <td>184.16</td>\n",
       "      <td>2136.70</td>\n",
       "      <td>95.89</td>\n",
       "      <td>149.58</td>\n",
       "      <td>204.72</td>\n",
       "      <td>282.58</td>\n",
       "      <td>114.22</td>\n",
       "      <td>235.415</td>\n",
       "      <td>164.832</td>\n",
       "      <td>...</td>\n",
       "      <td>140.31</td>\n",
       "      <td>186.850</td>\n",
       "      <td>132.90</td>\n",
       "      <td>70.49</td>\n",
       "      <td>133.92</td>\n",
       "      <td>70.48</td>\n",
       "      <td>154.12</td>\n",
       "      <td>466.205</td>\n",
       "      <td>130.183333</td>\n",
       "      <td>169.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391195</th>\n",
       "      <td>1.0</td>\n",
       "      <td>186.55</td>\n",
       "      <td>1987.63</td>\n",
       "      <td>95.84</td>\n",
       "      <td>151.39</td>\n",
       "      <td>188.04</td>\n",
       "      <td>252.95</td>\n",
       "      <td>117.16</td>\n",
       "      <td>221.765</td>\n",
       "      <td>196.354</td>\n",
       "      <td>...</td>\n",
       "      <td>108.77</td>\n",
       "      <td>209.540</td>\n",
       "      <td>117.39</td>\n",
       "      <td>64.92</td>\n",
       "      <td>110.48</td>\n",
       "      <td>56.58</td>\n",
       "      <td>160.57</td>\n",
       "      <td>484.980</td>\n",
       "      <td>124.223333</td>\n",
       "      <td>178.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391196</th>\n",
       "      <td>1.0</td>\n",
       "      <td>204.18</td>\n",
       "      <td>1182.57</td>\n",
       "      <td>98.78</td>\n",
       "      <td>149.03</td>\n",
       "      <td>216.82</td>\n",
       "      <td>276.85</td>\n",
       "      <td>135.16</td>\n",
       "      <td>209.215</td>\n",
       "      <td>133.932</td>\n",
       "      <td>...</td>\n",
       "      <td>136.20</td>\n",
       "      <td>211.305</td>\n",
       "      <td>139.30</td>\n",
       "      <td>72.91</td>\n",
       "      <td>111.31</td>\n",
       "      <td>57.00</td>\n",
       "      <td>159.50</td>\n",
       "      <td>406.645</td>\n",
       "      <td>123.346667</td>\n",
       "      <td>125.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391197</th>\n",
       "      <td>1.0</td>\n",
       "      <td>208.02</td>\n",
       "      <td>665.23</td>\n",
       "      <td>92.76</td>\n",
       "      <td>183.64</td>\n",
       "      <td>217.65</td>\n",
       "      <td>268.78</td>\n",
       "      <td>128.02</td>\n",
       "      <td>190.850</td>\n",
       "      <td>130.122</td>\n",
       "      <td>...</td>\n",
       "      <td>159.44</td>\n",
       "      <td>164.295</td>\n",
       "      <td>158.91</td>\n",
       "      <td>63.49</td>\n",
       "      <td>122.13</td>\n",
       "      <td>77.04</td>\n",
       "      <td>196.24</td>\n",
       "      <td>327.050</td>\n",
       "      <td>117.470000</td>\n",
       "      <td>321.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM391198</th>\n",
       "      <td>1.0</td>\n",
       "      <td>181.88</td>\n",
       "      <td>954.21</td>\n",
       "      <td>96.77</td>\n",
       "      <td>166.22</td>\n",
       "      <td>204.09</td>\n",
       "      <td>296.50</td>\n",
       "      <td>129.04</td>\n",
       "      <td>195.305</td>\n",
       "      <td>128.136</td>\n",
       "      <td>...</td>\n",
       "      <td>136.41</td>\n",
       "      <td>135.200</td>\n",
       "      <td>137.23</td>\n",
       "      <td>71.17</td>\n",
       "      <td>118.96</td>\n",
       "      <td>76.12</td>\n",
       "      <td>145.63</td>\n",
       "      <td>349.415</td>\n",
       "      <td>123.816667</td>\n",
       "      <td>187.64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 11728 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Kidney Chromophobe    A1CF      A2M  A4GALT   A4GNT    AAAS  \\\n",
       "GSM391107                 1.0  216.85  1234.58  156.47  114.73  205.97   \n",
       "GSM391108                 1.0  273.32  1600.46  147.93   96.43  161.96   \n",
       "GSM391109                 1.0  286.00  1138.55  250.03  115.99  186.94   \n",
       "GSM391110                 1.0  270.79  1222.28  209.76   82.23  175.21   \n",
       "GSM391111                 1.0  234.91  1633.94  244.66   91.99  169.83   \n",
       "...                       ...     ...      ...     ...     ...     ...   \n",
       "GSM391194                 1.0  184.16  2136.70   95.89  149.58  204.72   \n",
       "GSM391195                 1.0  186.55  1987.63   95.84  151.39  188.04   \n",
       "GSM391196                 1.0  204.18  1182.57   98.78  149.03  216.82   \n",
       "GSM391197                 1.0  208.02   665.23   92.76  183.64  217.65   \n",
       "GSM391198                 1.0  181.88   954.21   96.77  166.22  204.09   \n",
       "\n",
       "             AACS   AADAC    AAGAB     AAK1  ...  ZSWIM1   ZSWIM8    ZW10  \\\n",
       "GSM391107  275.87   89.78  166.720  181.854  ...  139.83  136.890  164.60   \n",
       "GSM391108  237.88   74.43  202.045  166.652  ...  137.27  127.165  131.09   \n",
       "GSM391109  243.51   72.76  204.975  149.652  ...  129.74  175.435  142.49   \n",
       "GSM391110  187.69   72.96  161.780  151.998  ...  108.38  171.055  166.73   \n",
       "GSM391111  216.02   68.80  172.450  129.822  ...   95.99  150.990  159.86   \n",
       "...           ...     ...      ...      ...  ...     ...      ...     ...   \n",
       "GSM391194  282.58  114.22  235.415  164.832  ...  140.31  186.850  132.90   \n",
       "GSM391195  252.95  117.16  221.765  196.354  ...  108.77  209.540  117.39   \n",
       "GSM391196  276.85  135.16  209.215  133.932  ...  136.20  211.305  139.30   \n",
       "GSM391197  268.78  128.02  190.850  130.122  ...  159.44  164.295  158.91   \n",
       "GSM391198  296.50  129.04  195.305  128.136  ...  136.41  135.200  137.23   \n",
       "\n",
       "           ZWILCH   ZWINT   ZXDB    ZXDC      ZYX       ZZEF1    ZZZ3  \n",
       "GSM391107   88.83  124.41  76.15  229.67  364.805  110.290000  120.57  \n",
       "GSM391108   77.84  164.05  73.69  204.86  319.760  115.060000  168.53  \n",
       "GSM391109   50.41  112.01  67.46  171.43  403.445  104.916667  115.28  \n",
       "GSM391110   54.93   80.23  59.64  198.63  459.975   96.573333  149.22  \n",
       "GSM391111   76.54  100.93  69.45  208.86  665.935   95.296667  149.95  \n",
       "...           ...     ...    ...     ...      ...         ...     ...  \n",
       "GSM391194   70.49  133.92  70.48  154.12  466.205  130.183333  169.52  \n",
       "GSM391195   64.92  110.48  56.58  160.57  484.980  124.223333  178.07  \n",
       "GSM391196   72.91  111.31  57.00  159.50  406.645  123.346667  125.93  \n",
       "GSM391197   63.49  122.13  77.04  196.24  327.050  117.470000  321.24  \n",
       "GSM391198   71.17  118.96  76.12  145.63  349.415  123.816667  187.64  \n",
       "\n",
       "[92 rows x 11728 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 92 samples.\n",
      "For the feature 'Kidney Chromophobe', the least common label is '1.0' with 92 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/zhangmianchen/Downloads/DATA/GEO/Kidney-Chromophobe/GSE17746/GSE17746_family.soft.gz',\n",
       " '/Users/zhangmianchen/Downloads/DATA/GEO/Kidney-Chromophobe/GSE17746/GSE17746-GPL9070_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = accession_num = \"GSE17746\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Sequential Binary Gene-Ratio Tests Define a Novel Molecular Diagnostic Strategy for Malignant Pleural Mesothelioma\"\n",
      "!Series_summary\t\"The gene-expression ratio  technique was used to design a molecular signature to diagnose MPM from among other potentially confounding diagnoses and differentiate the epithelioid from the sarcomatoid histological subtype of MPM.\"\n",
      "!Series_overall_design\t\"Microarray analysis was performed on 113 specimens including MPMs and a spectrum of tumors and benign tissues comprising the differential diagnosis of MPM.  A sequential combination of binary gene-expression ratio tests was developed to discriminate MPM from other thoracic malignancies .  This method was compared to other bioinformatic tools and this signature was validated  in an independent set of 170 samples.  Functional enrichment analysis was performed to identify differentially expressed probes.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1054230</th>\n",
       "      <th>GSM1054231</th>\n",
       "      <th>GSM1054232</th>\n",
       "      <th>GSM1054233</th>\n",
       "      <th>GSM1054234</th>\n",
       "      <th>GSM1054235</th>\n",
       "      <th>GSM1054236</th>\n",
       "      <th>GSM1054237</th>\n",
       "      <th>GSM1054238</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1054337</th>\n",
       "      <th>GSM1054338</th>\n",
       "      <th>GSM1054339</th>\n",
       "      <th>GSM1054340</th>\n",
       "      <th>GSM1054341</th>\n",
       "      <th>GSM1054342</th>\n",
       "      <th>GSM1054343</th>\n",
       "      <th>GSM1054344</th>\n",
       "      <th>GSM1054345</th>\n",
       "      <th>GSM1054346</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: control</td>\n",
       "      <td>tissue: control</td>\n",
       "      <td>tissue: control</td>\n",
       "      <td>tissue: control</td>\n",
       "      <td>tissue: Spindle Cell Sarcoma</td>\n",
       "      <td>tissue: Sarcoma</td>\n",
       "      <td>tissue: Sarcoma</td>\n",
       "      <td>tissue: Metastatic Melanoma</td>\n",
       "      <td>tissue: Pleomorphic Sarcoma</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Normal Pleura</td>\n",
       "      <td>tissue: Normal Pleura</td>\n",
       "      <td>tissue: Normal Pleura</td>\n",
       "      <td>tissue: Normal Pleura</td>\n",
       "      <td>tissue: MPM Sarcomatoid</td>\n",
       "      <td>tissue: MPM Biphasic</td>\n",
       "      <td>tissue: MPM Biphasic</td>\n",
       "      <td>tissue: MPM Biphasic</td>\n",
       "      <td>tissue: MPM Biphasic</td>\n",
       "      <td>tissue: MPM Biphasic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 118 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession       GSM1054230       GSM1054231  \\\n",
       "0  !Sample_characteristics_ch1  tissue: control  tissue: control   \n",
       "\n",
       "        GSM1054232       GSM1054233                    GSM1054234  \\\n",
       "0  tissue: control  tissue: control  tissue: Spindle Cell Sarcoma   \n",
       "\n",
       "        GSM1054235       GSM1054236                   GSM1054237  \\\n",
       "0  tissue: Sarcoma  tissue: Sarcoma  tissue: Metastatic Melanoma   \n",
       "\n",
       "                    GSM1054238  ...             GSM1054337  \\\n",
       "0  tissue: Pleomorphic Sarcoma  ...  tissue: Normal Pleura   \n",
       "\n",
       "              GSM1054338             GSM1054339             GSM1054340  \\\n",
       "0  tissue: Normal Pleura  tissue: Normal Pleura  tissue: Normal Pleura   \n",
       "\n",
       "                GSM1054341            GSM1054342            GSM1054343  \\\n",
       "0  tissue: MPM Sarcomatoid  tissue: MPM Biphasic  tissue: MPM Biphasic   \n",
       "\n",
       "             GSM1054344            GSM1054345            GSM1054346  \n",
       "0  tissue: MPM Biphasic  tissue: MPM Biphasic  tissue: MPM Biphasic  \n",
       "\n",
       "[1 rows x 118 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE42977\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'tissue: control',\n",
       "       'tissue: Spindle Cell Sarcoma', 'tissue: Sarcoma',\n",
       "       'tissue: Metastatic Melanoma', 'tissue: Pleomorphic Sarcoma',\n",
       "       'tissue: Renal Cell Carcinoma-Clear Cell',\n",
       "       'tissue: Synovial Sarcoma', 'tissue: Metastatic Thymoma',\n",
       "       'tissue: Metastatic Prostate Cancer',\n",
       "       'tissue: Stomach Cancer-Stromal Sarcoma',\n",
       "       'tissue: Non-Hodgkins Lymphoma', 'tissue: Hemangioendothelioma',\n",
       "       'tissue: Papillary Thyroid Carcinoma',\n",
       "       'tissue: Metastatic Thyroid Cancer',\n",
       "       'tissue: Lymphocytic Lymphoma', 'tissue: Thymoma',\n",
       "       'tissue: Melanoma-Malignant', 'tissue: Hemangiopericytoma',\n",
       "       'tissue: Thyroid Carcinoma', 'tissue: Monophasic Synovial Sarcoma',\n",
       "       'tissue: Metastatic Alveolar Soft Part Sarcoma',\n",
       "       'tissue: Metastatic Meningeal Hemangiopericytoma',\n",
       "       'tissue: Follicular Lymphoma', 'tissue: Rhabdomyosarcoma',\n",
       "       'tissue: Myofibrosarcoma',\n",
       "       'tissue: Renal Cell Carcinoma - Chromophobe',\n",
       "       'tissue: MPM Epithelioid', 'tissue: MPM Sarcomatoid',\n",
       "       'tissue: MPM Biphasic', 'tissue: Normal Lung',\n",
       "       'tissue: Lung  Adenocarcinoma',\n",
       "       'tissue: Metastatic Leiomyosarcoma', 'tissue: Colon Normal',\n",
       "       'tissue: K2/E5 met.prostate', 'tissue: Metastatic Chondrosarcoma',\n",
       "       'tissue: Metastatic Angiosarcoma', 'tissue: Normal Pleura'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Renal Cell Carcinoma - Chromophobe':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1054230</th>\n",
       "      <th>GSM1054231</th>\n",
       "      <th>GSM1054232</th>\n",
       "      <th>GSM1054233</th>\n",
       "      <th>GSM1054234</th>\n",
       "      <th>GSM1054235</th>\n",
       "      <th>GSM1054236</th>\n",
       "      <th>GSM1054237</th>\n",
       "      <th>GSM1054238</th>\n",
       "      <th>GSM1054239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1054337</th>\n",
       "      <th>GSM1054338</th>\n",
       "      <th>GSM1054339</th>\n",
       "      <th>GSM1054340</th>\n",
       "      <th>GSM1054341</th>\n",
       "      <th>GSM1054342</th>\n",
       "      <th>GSM1054343</th>\n",
       "      <th>GSM1054344</th>\n",
       "      <th>GSM1054345</th>\n",
       "      <th>GSM1054346</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 117 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM1054230  GSM1054231  GSM1054232  GSM1054233  \\\n",
       "Kidney Chromophobe           0           0           0           0   \n",
       "\n",
       "                    GSM1054234  GSM1054235  GSM1054236  GSM1054237  \\\n",
       "Kidney Chromophobe           0           0           0           0   \n",
       "\n",
       "                    GSM1054238  GSM1054239  ...  GSM1054337  GSM1054338  \\\n",
       "Kidney Chromophobe           0           0  ...           0           0   \n",
       "\n",
       "                    GSM1054339  GSM1054340  GSM1054341  GSM1054342  \\\n",
       "Kidney Chromophobe           0           0           0           0   \n",
       "\n",
       "                    GSM1054343  GSM1054344  GSM1054345  GSM1054346  \n",
       "Kidney Chromophobe           0           0           0           0  \n",
       "\n",
       "[1 rows x 117 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1054230</th>\n",
       "      <th>GSM1054231</th>\n",
       "      <th>GSM1054232</th>\n",
       "      <th>GSM1054233</th>\n",
       "      <th>GSM1054234</th>\n",
       "      <th>GSM1054235</th>\n",
       "      <th>GSM1054236</th>\n",
       "      <th>GSM1054237</th>\n",
       "      <th>GSM1054238</th>\n",
       "      <th>GSM1054239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1054337</th>\n",
       "      <th>GSM1054338</th>\n",
       "      <th>GSM1054339</th>\n",
       "      <th>GSM1054340</th>\n",
       "      <th>GSM1054341</th>\n",
       "      <th>GSM1054342</th>\n",
       "      <th>GSM1054343</th>\n",
       "      <th>GSM1054344</th>\n",
       "      <th>GSM1054345</th>\n",
       "      <th>GSM1054346</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_10000</th>\n",
       "      <td>1.972137</td>\n",
       "      <td>1.511342</td>\n",
       "      <td>2.845652</td>\n",
       "      <td>1.183206</td>\n",
       "      <td>2.740848</td>\n",
       "      <td>3.472879</td>\n",
       "      <td>3.876484</td>\n",
       "      <td>7.543180</td>\n",
       "      <td>3.854739</td>\n",
       "      <td>1.886201</td>\n",
       "      <td>...</td>\n",
       "      <td>2.208187</td>\n",
       "      <td>1.678540</td>\n",
       "      <td>2.347179</td>\n",
       "      <td>2.308059</td>\n",
       "      <td>4.152523</td>\n",
       "      <td>4.434662</td>\n",
       "      <td>5.518134</td>\n",
       "      <td>15.222088</td>\n",
       "      <td>8.603092</td>\n",
       "      <td>4.540210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_100000</th>\n",
       "      <td>0.924305</td>\n",
       "      <td>0.812420</td>\n",
       "      <td>0.937917</td>\n",
       "      <td>0.942640</td>\n",
       "      <td>0.916636</td>\n",
       "      <td>0.870865</td>\n",
       "      <td>0.775436</td>\n",
       "      <td>0.872982</td>\n",
       "      <td>0.897949</td>\n",
       "      <td>0.925115</td>\n",
       "      <td>...</td>\n",
       "      <td>0.824272</td>\n",
       "      <td>0.884153</td>\n",
       "      <td>0.874214</td>\n",
       "      <td>0.856160</td>\n",
       "      <td>0.943487</td>\n",
       "      <td>0.948834</td>\n",
       "      <td>0.869908</td>\n",
       "      <td>0.857906</td>\n",
       "      <td>0.757040</td>\n",
       "      <td>0.872259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_100007</th>\n",
       "      <td>0.909786</td>\n",
       "      <td>0.795458</td>\n",
       "      <td>0.936011</td>\n",
       "      <td>0.857263</td>\n",
       "      <td>0.870995</td>\n",
       "      <td>0.775896</td>\n",
       "      <td>0.816944</td>\n",
       "      <td>0.830887</td>\n",
       "      <td>0.811600</td>\n",
       "      <td>0.820598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.711004</td>\n",
       "      <td>0.835415</td>\n",
       "      <td>0.913357</td>\n",
       "      <td>0.890036</td>\n",
       "      <td>0.805675</td>\n",
       "      <td>0.930804</td>\n",
       "      <td>0.894138</td>\n",
       "      <td>0.817027</td>\n",
       "      <td>0.977993</td>\n",
       "      <td>0.944993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_100009</th>\n",
       "      <td>0.883831</td>\n",
       "      <td>0.857333</td>\n",
       "      <td>0.811424</td>\n",
       "      <td>0.907362</td>\n",
       "      <td>0.763149</td>\n",
       "      <td>0.764961</td>\n",
       "      <td>0.845055</td>\n",
       "      <td>0.898741</td>\n",
       "      <td>0.932021</td>\n",
       "      <td>0.951832</td>\n",
       "      <td>...</td>\n",
       "      <td>0.779348</td>\n",
       "      <td>0.912258</td>\n",
       "      <td>0.887889</td>\n",
       "      <td>0.879680</td>\n",
       "      <td>0.869804</td>\n",
       "      <td>0.685800</td>\n",
       "      <td>0.786594</td>\n",
       "      <td>0.752266</td>\n",
       "      <td>0.791208</td>\n",
       "      <td>0.868821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_10001</th>\n",
       "      <td>42.180716</td>\n",
       "      <td>28.295220</td>\n",
       "      <td>48.491675</td>\n",
       "      <td>15.108901</td>\n",
       "      <td>50.917271</td>\n",
       "      <td>12.615383</td>\n",
       "      <td>32.558544</td>\n",
       "      <td>48.668974</td>\n",
       "      <td>37.672912</td>\n",
       "      <td>16.421997</td>\n",
       "      <td>...</td>\n",
       "      <td>62.377845</td>\n",
       "      <td>71.203787</td>\n",
       "      <td>21.897301</td>\n",
       "      <td>29.824924</td>\n",
       "      <td>16.600328</td>\n",
       "      <td>15.631902</td>\n",
       "      <td>16.028539</td>\n",
       "      <td>23.609221</td>\n",
       "      <td>16.012058</td>\n",
       "      <td>36.767564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_99987</th>\n",
       "      <td>0.734585</td>\n",
       "      <td>0.818516</td>\n",
       "      <td>0.767362</td>\n",
       "      <td>0.877341</td>\n",
       "      <td>0.790919</td>\n",
       "      <td>0.884766</td>\n",
       "      <td>0.762665</td>\n",
       "      <td>0.930539</td>\n",
       "      <td>0.877803</td>\n",
       "      <td>0.897816</td>\n",
       "      <td>...</td>\n",
       "      <td>0.850284</td>\n",
       "      <td>0.895458</td>\n",
       "      <td>0.845082</td>\n",
       "      <td>0.817136</td>\n",
       "      <td>0.811308</td>\n",
       "      <td>0.863365</td>\n",
       "      <td>0.797534</td>\n",
       "      <td>0.797442</td>\n",
       "      <td>0.836796</td>\n",
       "      <td>0.831163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_9999</th>\n",
       "      <td>2.186225</td>\n",
       "      <td>2.062302</td>\n",
       "      <td>1.901461</td>\n",
       "      <td>1.265443</td>\n",
       "      <td>1.524241</td>\n",
       "      <td>1.100508</td>\n",
       "      <td>2.845723</td>\n",
       "      <td>1.751512</td>\n",
       "      <td>1.252493</td>\n",
       "      <td>1.431054</td>\n",
       "      <td>...</td>\n",
       "      <td>2.442431</td>\n",
       "      <td>1.699486</td>\n",
       "      <td>1.168026</td>\n",
       "      <td>1.244256</td>\n",
       "      <td>1.859018</td>\n",
       "      <td>1.739583</td>\n",
       "      <td>1.604486</td>\n",
       "      <td>1.295576</td>\n",
       "      <td>1.220029</td>\n",
       "      <td>2.120357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_99990</th>\n",
       "      <td>1.003755</td>\n",
       "      <td>1.105511</td>\n",
       "      <td>0.912209</td>\n",
       "      <td>1.125055</td>\n",
       "      <td>1.005835</td>\n",
       "      <td>1.019736</td>\n",
       "      <td>1.077536</td>\n",
       "      <td>0.985023</td>\n",
       "      <td>1.006931</td>\n",
       "      <td>0.959392</td>\n",
       "      <td>...</td>\n",
       "      <td>1.095790</td>\n",
       "      <td>1.086057</td>\n",
       "      <td>1.050433</td>\n",
       "      <td>0.964199</td>\n",
       "      <td>1.093784</td>\n",
       "      <td>1.000962</td>\n",
       "      <td>0.997990</td>\n",
       "      <td>1.111851</td>\n",
       "      <td>0.976525</td>\n",
       "      <td>1.035464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_99995</th>\n",
       "      <td>0.867995</td>\n",
       "      <td>0.913600</td>\n",
       "      <td>1.069407</td>\n",
       "      <td>0.840991</td>\n",
       "      <td>0.966751</td>\n",
       "      <td>0.978569</td>\n",
       "      <td>0.851452</td>\n",
       "      <td>0.917244</td>\n",
       "      <td>0.944750</td>\n",
       "      <td>0.969894</td>\n",
       "      <td>...</td>\n",
       "      <td>0.879109</td>\n",
       "      <td>0.928430</td>\n",
       "      <td>0.867635</td>\n",
       "      <td>1.020188</td>\n",
       "      <td>0.820790</td>\n",
       "      <td>0.957369</td>\n",
       "      <td>0.846994</td>\n",
       "      <td>0.922516</td>\n",
       "      <td>0.958226</td>\n",
       "      <td>0.915698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_99996</th>\n",
       "      <td>0.963215</td>\n",
       "      <td>0.796494</td>\n",
       "      <td>0.861250</td>\n",
       "      <td>0.992920</td>\n",
       "      <td>0.793244</td>\n",
       "      <td>0.859421</td>\n",
       "      <td>0.890161</td>\n",
       "      <td>0.850599</td>\n",
       "      <td>0.885614</td>\n",
       "      <td>0.864726</td>\n",
       "      <td>...</td>\n",
       "      <td>0.876516</td>\n",
       "      <td>0.873626</td>\n",
       "      <td>0.995070</td>\n",
       "      <td>0.961808</td>\n",
       "      <td>0.944668</td>\n",
       "      <td>0.845995</td>\n",
       "      <td>0.863367</td>\n",
       "      <td>0.827960</td>\n",
       "      <td>0.879339</td>\n",
       "      <td>0.879238</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46713 rows × 117 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             GSM1054230  GSM1054231  GSM1054232  GSM1054233  GSM1054234  \\\n",
       "ID                                                                        \n",
       "ILMN_10000     1.972137    1.511342    2.845652    1.183206    2.740848   \n",
       "ILMN_100000    0.924305    0.812420    0.937917    0.942640    0.916636   \n",
       "ILMN_100007    0.909786    0.795458    0.936011    0.857263    0.870995   \n",
       "ILMN_100009    0.883831    0.857333    0.811424    0.907362    0.763149   \n",
       "ILMN_10001    42.180716   28.295220   48.491675   15.108901   50.917271   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "ILMN_99987     0.734585    0.818516    0.767362    0.877341    0.790919   \n",
       "ILMN_9999      2.186225    2.062302    1.901461    1.265443    1.524241   \n",
       "ILMN_99990     1.003755    1.105511    0.912209    1.125055    1.005835   \n",
       "ILMN_99995     0.867995    0.913600    1.069407    0.840991    0.966751   \n",
       "ILMN_99996     0.963215    0.796494    0.861250    0.992920    0.793244   \n",
       "\n",
       "             GSM1054235  GSM1054236  GSM1054237  GSM1054238  GSM1054239  ...  \\\n",
       "ID                                                                       ...   \n",
       "ILMN_10000     3.472879    3.876484    7.543180    3.854739    1.886201  ...   \n",
       "ILMN_100000    0.870865    0.775436    0.872982    0.897949    0.925115  ...   \n",
       "ILMN_100007    0.775896    0.816944    0.830887    0.811600    0.820598  ...   \n",
       "ILMN_100009    0.764961    0.845055    0.898741    0.932021    0.951832  ...   \n",
       "ILMN_10001    12.615383   32.558544   48.668974   37.672912   16.421997  ...   \n",
       "...                 ...         ...         ...         ...         ...  ...   \n",
       "ILMN_99987     0.884766    0.762665    0.930539    0.877803    0.897816  ...   \n",
       "ILMN_9999      1.100508    2.845723    1.751512    1.252493    1.431054  ...   \n",
       "ILMN_99990     1.019736    1.077536    0.985023    1.006931    0.959392  ...   \n",
       "ILMN_99995     0.978569    0.851452    0.917244    0.944750    0.969894  ...   \n",
       "ILMN_99996     0.859421    0.890161    0.850599    0.885614    0.864726  ...   \n",
       "\n",
       "             GSM1054337  GSM1054338  GSM1054339  GSM1054340  GSM1054341  \\\n",
       "ID                                                                        \n",
       "ILMN_10000     2.208187    1.678540    2.347179    2.308059    4.152523   \n",
       "ILMN_100000    0.824272    0.884153    0.874214    0.856160    0.943487   \n",
       "ILMN_100007    0.711004    0.835415    0.913357    0.890036    0.805675   \n",
       "ILMN_100009    0.779348    0.912258    0.887889    0.879680    0.869804   \n",
       "ILMN_10001    62.377845   71.203787   21.897301   29.824924   16.600328   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "ILMN_99987     0.850284    0.895458    0.845082    0.817136    0.811308   \n",
       "ILMN_9999      2.442431    1.699486    1.168026    1.244256    1.859018   \n",
       "ILMN_99990     1.095790    1.086057    1.050433    0.964199    1.093784   \n",
       "ILMN_99995     0.879109    0.928430    0.867635    1.020188    0.820790   \n",
       "ILMN_99996     0.876516    0.873626    0.995070    0.961808    0.944668   \n",
       "\n",
       "             GSM1054342  GSM1054343  GSM1054344  GSM1054345  GSM1054346  \n",
       "ID                                                                       \n",
       "ILMN_10000     4.434662    5.518134   15.222088    8.603092    4.540210  \n",
       "ILMN_100000    0.948834    0.869908    0.857906    0.757040    0.872259  \n",
       "ILMN_100007    0.930804    0.894138    0.817027    0.977993    0.944993  \n",
       "ILMN_100009    0.685800    0.786594    0.752266    0.791208    0.868821  \n",
       "ILMN_10001    15.631902   16.028539   23.609221   16.012058   36.767564  \n",
       "...                 ...         ...         ...         ...         ...  \n",
       "ILMN_99987     0.863365    0.797534    0.797442    0.836796    0.831163  \n",
       "ILMN_9999      1.739583    1.604486    1.295576    1.220029    2.120357  \n",
       "ILMN_99990     1.000962    0.997990    1.111851    0.976525    1.035464  \n",
       "ILMN_99995     0.957369    0.846994    0.922516    0.958226    0.915698  \n",
       "ILMN_99996     0.845995    0.863367    0.827960    0.879339    0.879238  \n",
       "\n",
       "[46713 rows x 117 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_89282', 'ILMN_35826', 'ILMN_25544', 'ILMN_132331', 'ILMN_105017'], 'GB_ACC': ['BU678343', 'XM_497527.2', 'NM_018433.3', 'AW629334', 'AI818233'], 'Symbol': [nan, 'LOC441782', 'JMJD1A', nan, nan], 'SEQUENCE': ['CTCTCTAAAGGGACAACAGAGTGGACAGTCAAGGAACTCCACATATTCAT', 'GGGGTCAAGCCCAGGTGAAATGTGGATTGGAAAAGTGCTTCCCTTGCCCC', 'CCAGGCTGTAAAAGCAAAACCTCGTATCAGCTCTGGAACAATACCTGCAG', 'CCAGACAGGAAGCATCAAGCCCTTCAGGAAAGAATATGCGAGAGTGCTGC', 'TGTGCAGAAAGCTGATGGAAGGGAGAAAGAATGGAAGTGGGTCACACAGC'], 'Definition': ['UI-CF-EC0-abi-c-12-0-UI.s1 UI-CF-EC0 Homo sapiens cDNA clone UI-CF-EC0-abi-c-12-0-UI 3, mRNA sequence', 'PREDICTED: Homo sapiens similar to spectrin domain with coiled-coils 1 (LOC441782), mRNA.', 'Homo sapiens jumonji domain containing 1A (JMJD1A), mRNA.', 'hi56g05.x1 Soares_NFL_T_GBC_S1 Homo sapiens cDNA clone IMAGE:2976344 3, mRNA sequence', 'wk77d04.x1 NCI_CGAP_Pan1 Homo sapiens cDNA clone IMAGE:2421415 3, mRNA sequence'], 'Ontology': [nan, nan, nan, nan, nan], 'Synonym': [nan, nan, 'TSGA; JMJD1; KIAA0742; DKFZp686A24246; DKFZp686P07111', nan, nan]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['ILMN_10000', 'ILMN_100000', 'ILMN_100007', 'ILMN_100009', 'ILMN_10001', 'ILMN_100010', 'ILMN_10002', 'ILMN_100028', 'ILMN_100030', 'ILMN_100031', 'ILMN_100034', 'ILMN_100037', 'ILMN_10004', 'ILMN_10005', 'ILMN_100054', 'ILMN_100059', 'ILMN_10006', 'ILMN_100075', 'ILMN_100079', 'ILMN_100083']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_89282', 'ILMN_35826', 'ILMN_25544', 'ILMN_132331', 'ILMN_105017'], 'GB_ACC': ['BU678343', 'XM_497527.2', 'NM_018433.3', 'AW629334', 'AI818233'], 'Symbol': [nan, 'LOC441782', 'JMJD1A', nan, nan], 'SEQUENCE': ['CTCTCTAAAGGGACAACAGAGTGGACAGTCAAGGAACTCCACATATTCAT', 'GGGGTCAAGCCCAGGTGAAATGTGGATTGGAAAAGTGCTTCCCTTGCCCC', 'CCAGGCTGTAAAAGCAAAACCTCGTATCAGCTCTGGAACAATACCTGCAG', 'CCAGACAGGAAGCATCAAGCCCTTCAGGAAAGAATATGCGAGAGTGCTGC', 'TGTGCAGAAAGCTGATGGAAGGGAGAAAGAATGGAAGTGGGTCACACAGC'], 'Definition': ['UI-CF-EC0-abi-c-12-0-UI.s1 UI-CF-EC0 Homo sapiens cDNA clone UI-CF-EC0-abi-c-12-0-UI 3, mRNA sequence', 'PREDICTED: Homo sapiens similar to spectrin domain with coiled-coils 1 (LOC441782), mRNA.', 'Homo sapiens jumonji domain containing 1A (JMJD1A), mRNA.', 'hi56g05.x1 Soares_NFL_T_GBC_S1 Homo sapiens cDNA clone IMAGE:2976344 3, mRNA sequence', 'wk77d04.x1 NCI_CGAP_Pan1 Homo sapiens cDNA clone IMAGE:2421415 3, mRNA sequence'], 'Ontology': [nan, nan, nan, nan, nan], 'Synonym': [nan, nan, 'TSGA; JMJD1; KIAA0742; DKFZp686A24246; DKFZp686P07111', nan, nan]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'Symbol', 'SEQUENCE', 'Definition', 'Ontology',\n",
       "       'Synonym'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1054230</th>\n",
       "      <th>GSM1054231</th>\n",
       "      <th>GSM1054232</th>\n",
       "      <th>GSM1054233</th>\n",
       "      <th>GSM1054234</th>\n",
       "      <th>GSM1054235</th>\n",
       "      <th>GSM1054236</th>\n",
       "      <th>GSM1054237</th>\n",
       "      <th>GSM1054238</th>\n",
       "      <th>GSM1054239</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1054337</th>\n",
       "      <th>GSM1054338</th>\n",
       "      <th>GSM1054339</th>\n",
       "      <th>GSM1054340</th>\n",
       "      <th>GSM1054341</th>\n",
       "      <th>GSM1054342</th>\n",
       "      <th>GSM1054343</th>\n",
       "      <th>GSM1054344</th>\n",
       "      <th>GSM1054345</th>\n",
       "      <th>GSM1054346</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.903276</td>\n",
       "      <td>0.821580</td>\n",
       "      <td>0.865428</td>\n",
       "      <td>0.890772</td>\n",
       "      <td>0.986434</td>\n",
       "      <td>1.197420</td>\n",
       "      <td>0.886271</td>\n",
       "      <td>1.117455</td>\n",
       "      <td>1.003215</td>\n",
       "      <td>0.865700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.855990</td>\n",
       "      <td>0.897686</td>\n",
       "      <td>0.884723</td>\n",
       "      <td>0.925664</td>\n",
       "      <td>1.061455</td>\n",
       "      <td>0.868972</td>\n",
       "      <td>0.961211</td>\n",
       "      <td>0.802412</td>\n",
       "      <td>0.964033</td>\n",
       "      <td>1.097807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>27.154642</td>\n",
       "      <td>17.070914</td>\n",
       "      <td>34.490890</td>\n",
       "      <td>9.773921</td>\n",
       "      <td>31.282731</td>\n",
       "      <td>20.569610</td>\n",
       "      <td>65.495478</td>\n",
       "      <td>77.842674</td>\n",
       "      <td>30.995838</td>\n",
       "      <td>41.236381</td>\n",
       "      <td>...</td>\n",
       "      <td>35.781895</td>\n",
       "      <td>43.580901</td>\n",
       "      <td>32.651689</td>\n",
       "      <td>25.655624</td>\n",
       "      <td>28.996303</td>\n",
       "      <td>22.165116</td>\n",
       "      <td>21.972497</td>\n",
       "      <td>22.618258</td>\n",
       "      <td>22.302836</td>\n",
       "      <td>34.703123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.893226</td>\n",
       "      <td>0.963602</td>\n",
       "      <td>0.996714</td>\n",
       "      <td>1.108153</td>\n",
       "      <td>0.855068</td>\n",
       "      <td>1.076572</td>\n",
       "      <td>0.943271</td>\n",
       "      <td>1.017992</td>\n",
       "      <td>0.925794</td>\n",
       "      <td>0.964499</td>\n",
       "      <td>...</td>\n",
       "      <td>1.070994</td>\n",
       "      <td>1.050752</td>\n",
       "      <td>1.000574</td>\n",
       "      <td>1.029542</td>\n",
       "      <td>1.056357</td>\n",
       "      <td>1.102940</td>\n",
       "      <td>0.976782</td>\n",
       "      <td>1.146317</td>\n",
       "      <td>1.128793</td>\n",
       "      <td>0.950305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>1.357179</td>\n",
       "      <td>1.423179</td>\n",
       "      <td>1.518340</td>\n",
       "      <td>1.352150</td>\n",
       "      <td>1.415984</td>\n",
       "      <td>1.592975</td>\n",
       "      <td>1.484345</td>\n",
       "      <td>1.446382</td>\n",
       "      <td>1.389944</td>\n",
       "      <td>1.314923</td>\n",
       "      <td>...</td>\n",
       "      <td>1.488361</td>\n",
       "      <td>1.419631</td>\n",
       "      <td>1.593990</td>\n",
       "      <td>1.414248</td>\n",
       "      <td>1.362292</td>\n",
       "      <td>1.279548</td>\n",
       "      <td>1.113990</td>\n",
       "      <td>1.113299</td>\n",
       "      <td>1.135271</td>\n",
       "      <td>1.320155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>2.008587</td>\n",
       "      <td>1.581049</td>\n",
       "      <td>2.094153</td>\n",
       "      <td>1.300390</td>\n",
       "      <td>11.355988</td>\n",
       "      <td>6.706682</td>\n",
       "      <td>7.649266</td>\n",
       "      <td>1.623299</td>\n",
       "      <td>2.566030</td>\n",
       "      <td>7.711680</td>\n",
       "      <td>...</td>\n",
       "      <td>2.302174</td>\n",
       "      <td>3.100832</td>\n",
       "      <td>2.500579</td>\n",
       "      <td>2.247663</td>\n",
       "      <td>4.225168</td>\n",
       "      <td>1.733522</td>\n",
       "      <td>4.810449</td>\n",
       "      <td>4.048876</td>\n",
       "      <td>2.373974</td>\n",
       "      <td>1.274872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>19.655399</td>\n",
       "      <td>15.911316</td>\n",
       "      <td>18.813045</td>\n",
       "      <td>9.908477</td>\n",
       "      <td>37.106860</td>\n",
       "      <td>32.012832</td>\n",
       "      <td>46.715755</td>\n",
       "      <td>25.140595</td>\n",
       "      <td>33.192002</td>\n",
       "      <td>11.446667</td>\n",
       "      <td>...</td>\n",
       "      <td>38.491333</td>\n",
       "      <td>31.592975</td>\n",
       "      <td>26.693239</td>\n",
       "      <td>21.626782</td>\n",
       "      <td>48.230134</td>\n",
       "      <td>33.685854</td>\n",
       "      <td>56.189855</td>\n",
       "      <td>58.047052</td>\n",
       "      <td>58.464519</td>\n",
       "      <td>60.301068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>7.226102</td>\n",
       "      <td>9.249937</td>\n",
       "      <td>7.148919</td>\n",
       "      <td>6.130382</td>\n",
       "      <td>9.664492</td>\n",
       "      <td>6.915493</td>\n",
       "      <td>11.620945</td>\n",
       "      <td>9.445520</td>\n",
       "      <td>6.164034</td>\n",
       "      <td>6.642245</td>\n",
       "      <td>...</td>\n",
       "      <td>10.193268</td>\n",
       "      <td>9.867647</td>\n",
       "      <td>6.958772</td>\n",
       "      <td>6.572962</td>\n",
       "      <td>7.505992</td>\n",
       "      <td>7.603477</td>\n",
       "      <td>9.829170</td>\n",
       "      <td>6.854160</td>\n",
       "      <td>9.044523</td>\n",
       "      <td>7.138543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>13.926936</td>\n",
       "      <td>8.321252</td>\n",
       "      <td>10.700753</td>\n",
       "      <td>4.641731</td>\n",
       "      <td>16.992292</td>\n",
       "      <td>6.677801</td>\n",
       "      <td>16.909833</td>\n",
       "      <td>13.829324</td>\n",
       "      <td>11.036007</td>\n",
       "      <td>6.636792</td>\n",
       "      <td>...</td>\n",
       "      <td>13.362898</td>\n",
       "      <td>11.496935</td>\n",
       "      <td>4.852836</td>\n",
       "      <td>5.190852</td>\n",
       "      <td>9.553233</td>\n",
       "      <td>11.981230</td>\n",
       "      <td>8.417731</td>\n",
       "      <td>7.689744</td>\n",
       "      <td>6.061908</td>\n",
       "      <td>14.245139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF2A</th>\n",
       "      <td>36.250133</td>\n",
       "      <td>21.393597</td>\n",
       "      <td>24.619330</td>\n",
       "      <td>10.641300</td>\n",
       "      <td>28.841662</td>\n",
       "      <td>15.424809</td>\n",
       "      <td>22.914809</td>\n",
       "      <td>28.124125</td>\n",
       "      <td>26.329730</td>\n",
       "      <td>15.381810</td>\n",
       "      <td>...</td>\n",
       "      <td>28.626723</td>\n",
       "      <td>33.113892</td>\n",
       "      <td>13.609499</td>\n",
       "      <td>14.907299</td>\n",
       "      <td>14.487761</td>\n",
       "      <td>16.448616</td>\n",
       "      <td>16.942906</td>\n",
       "      <td>25.439824</td>\n",
       "      <td>13.607890</td>\n",
       "      <td>31.558448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB1C</th>\n",
       "      <td>0.815878</td>\n",
       "      <td>0.698088</td>\n",
       "      <td>0.820355</td>\n",
       "      <td>0.759291</td>\n",
       "      <td>0.779781</td>\n",
       "      <td>0.853471</td>\n",
       "      <td>0.825560</td>\n",
       "      <td>0.823434</td>\n",
       "      <td>0.861270</td>\n",
       "      <td>0.803424</td>\n",
       "      <td>...</td>\n",
       "      <td>0.744824</td>\n",
       "      <td>0.853769</td>\n",
       "      <td>0.828589</td>\n",
       "      <td>0.833693</td>\n",
       "      <td>0.848270</td>\n",
       "      <td>1.010288</td>\n",
       "      <td>0.879999</td>\n",
       "      <td>1.039451</td>\n",
       "      <td>1.062880</td>\n",
       "      <td>0.901901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13857 rows × 117 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1054230  GSM1054231  GSM1054232  GSM1054233  GSM1054234  \\\n",
       "A1BG       0.903276    0.821580    0.865428    0.890772    0.986434   \n",
       "A2M       27.154642   17.070914   34.490890    9.773921   31.282731   \n",
       "A2ML1      0.893226    0.963602    0.996714    1.108153    0.855068   \n",
       "A3GALT2    1.357179    1.423179    1.518340    1.352150    1.415984   \n",
       "A4GALT     2.008587    1.581049    2.094153    1.300390   11.355988   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYX       19.655399   15.911316   18.813045    9.908477   37.106860   \n",
       "ZZEF1      7.226102    9.249937    7.148919    6.130382    9.664492   \n",
       "ZZZ3      13.926936    8.321252   10.700753    4.641731   16.992292   \n",
       "EIF2A     36.250133   21.393597   24.619330   10.641300   28.841662   \n",
       "RAB1C      0.815878    0.698088    0.820355    0.759291    0.779781   \n",
       "\n",
       "         GSM1054235  GSM1054236  GSM1054237  GSM1054238  GSM1054239  ...  \\\n",
       "A1BG       1.197420    0.886271    1.117455    1.003215    0.865700  ...   \n",
       "A2M       20.569610   65.495478   77.842674   30.995838   41.236381  ...   \n",
       "A2ML1      1.076572    0.943271    1.017992    0.925794    0.964499  ...   \n",
       "A3GALT2    1.592975    1.484345    1.446382    1.389944    1.314923  ...   \n",
       "A4GALT     6.706682    7.649266    1.623299    2.566030    7.711680  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZYX       32.012832   46.715755   25.140595   33.192002   11.446667  ...   \n",
       "ZZEF1      6.915493   11.620945    9.445520    6.164034    6.642245  ...   \n",
       "ZZZ3       6.677801   16.909833   13.829324   11.036007    6.636792  ...   \n",
       "EIF2A     15.424809   22.914809   28.124125   26.329730   15.381810  ...   \n",
       "RAB1C      0.853471    0.825560    0.823434    0.861270    0.803424  ...   \n",
       "\n",
       "         GSM1054337  GSM1054338  GSM1054339  GSM1054340  GSM1054341  \\\n",
       "A1BG       0.855990    0.897686    0.884723    0.925664    1.061455   \n",
       "A2M       35.781895   43.580901   32.651689   25.655624   28.996303   \n",
       "A2ML1      1.070994    1.050752    1.000574    1.029542    1.056357   \n",
       "A3GALT2    1.488361    1.419631    1.593990    1.414248    1.362292   \n",
       "A4GALT     2.302174    3.100832    2.500579    2.247663    4.225168   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZYX       38.491333   31.592975   26.693239   21.626782   48.230134   \n",
       "ZZEF1     10.193268    9.867647    6.958772    6.572962    7.505992   \n",
       "ZZZ3      13.362898   11.496935    4.852836    5.190852    9.553233   \n",
       "EIF2A     28.626723   33.113892   13.609499   14.907299   14.487761   \n",
       "RAB1C      0.744824    0.853769    0.828589    0.833693    0.848270   \n",
       "\n",
       "         GSM1054342  GSM1054343  GSM1054344  GSM1054345  GSM1054346  \n",
       "A1BG       0.868972    0.961211    0.802412    0.964033    1.097807  \n",
       "A2M       22.165116   21.972497   22.618258   22.302836   34.703123  \n",
       "A2ML1      1.102940    0.976782    1.146317    1.128793    0.950305  \n",
       "A3GALT2    1.279548    1.113990    1.113299    1.135271    1.320155  \n",
       "A4GALT     1.733522    4.810449    4.048876    2.373974    1.274872  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZYX       33.685854   56.189855   58.047052   58.464519   60.301068  \n",
       "ZZEF1      7.603477    9.829170    6.854160    9.044523    7.138543  \n",
       "ZZZ3      11.981230    8.417731    7.689744    6.061908   14.245139  \n",
       "EIF2A     16.448616   16.942906   25.439824   13.607890   31.558448  \n",
       "RAB1C      1.010288    0.879999    1.039451    1.062880    0.901901  \n",
       "\n",
       "[13857 rows x 117 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>EIF2A</th>\n",
       "      <th>RAB1C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1054230</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.903276</td>\n",
       "      <td>27.154642</td>\n",
       "      <td>0.893226</td>\n",
       "      <td>1.357179</td>\n",
       "      <td>2.008587</td>\n",
       "      <td>1.090213</td>\n",
       "      <td>1.113465</td>\n",
       "      <td>3.079174</td>\n",
       "      <td>4.593716</td>\n",
       "      <td>...</td>\n",
       "      <td>0.956936</td>\n",
       "      <td>5.579101</td>\n",
       "      <td>2.600057</td>\n",
       "      <td>1.052827</td>\n",
       "      <td>15.959863</td>\n",
       "      <td>19.655399</td>\n",
       "      <td>7.226102</td>\n",
       "      <td>13.926936</td>\n",
       "      <td>36.250133</td>\n",
       "      <td>0.815878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054231</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.821580</td>\n",
       "      <td>17.070914</td>\n",
       "      <td>0.963602</td>\n",
       "      <td>1.423179</td>\n",
       "      <td>1.581049</td>\n",
       "      <td>1.108882</td>\n",
       "      <td>0.962892</td>\n",
       "      <td>1.792671</td>\n",
       "      <td>8.201332</td>\n",
       "      <td>...</td>\n",
       "      <td>0.911934</td>\n",
       "      <td>6.643073</td>\n",
       "      <td>1.840664</td>\n",
       "      <td>0.810640</td>\n",
       "      <td>53.815354</td>\n",
       "      <td>15.911316</td>\n",
       "      <td>9.249937</td>\n",
       "      <td>8.321252</td>\n",
       "      <td>21.393597</td>\n",
       "      <td>0.698088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054232</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.865428</td>\n",
       "      <td>34.490890</td>\n",
       "      <td>0.996714</td>\n",
       "      <td>1.518340</td>\n",
       "      <td>2.094153</td>\n",
       "      <td>1.133496</td>\n",
       "      <td>1.238555</td>\n",
       "      <td>3.890664</td>\n",
       "      <td>4.255689</td>\n",
       "      <td>...</td>\n",
       "      <td>0.868369</td>\n",
       "      <td>4.749234</td>\n",
       "      <td>2.945843</td>\n",
       "      <td>1.003140</td>\n",
       "      <td>15.712620</td>\n",
       "      <td>18.813045</td>\n",
       "      <td>7.148919</td>\n",
       "      <td>10.700753</td>\n",
       "      <td>24.619330</td>\n",
       "      <td>0.820355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054233</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.890772</td>\n",
       "      <td>9.773921</td>\n",
       "      <td>1.108153</td>\n",
       "      <td>1.352150</td>\n",
       "      <td>1.300390</td>\n",
       "      <td>1.041011</td>\n",
       "      <td>0.957591</td>\n",
       "      <td>1.664235</td>\n",
       "      <td>4.674827</td>\n",
       "      <td>...</td>\n",
       "      <td>0.892851</td>\n",
       "      <td>3.495437</td>\n",
       "      <td>1.402277</td>\n",
       "      <td>0.902783</td>\n",
       "      <td>31.327785</td>\n",
       "      <td>9.908477</td>\n",
       "      <td>6.130382</td>\n",
       "      <td>4.641731</td>\n",
       "      <td>10.641300</td>\n",
       "      <td>0.759291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054234</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.986434</td>\n",
       "      <td>31.282731</td>\n",
       "      <td>0.855068</td>\n",
       "      <td>1.415984</td>\n",
       "      <td>11.355988</td>\n",
       "      <td>1.010757</td>\n",
       "      <td>0.926479</td>\n",
       "      <td>2.094559</td>\n",
       "      <td>2.133348</td>\n",
       "      <td>...</td>\n",
       "      <td>0.953112</td>\n",
       "      <td>3.221408</td>\n",
       "      <td>2.098021</td>\n",
       "      <td>1.003471</td>\n",
       "      <td>24.561863</td>\n",
       "      <td>37.106860</td>\n",
       "      <td>9.664492</td>\n",
       "      <td>16.992292</td>\n",
       "      <td>28.841662</td>\n",
       "      <td>0.779781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054342</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.868972</td>\n",
       "      <td>22.165116</td>\n",
       "      <td>1.102940</td>\n",
       "      <td>1.279548</td>\n",
       "      <td>1.733522</td>\n",
       "      <td>1.089880</td>\n",
       "      <td>0.843426</td>\n",
       "      <td>1.531277</td>\n",
       "      <td>7.013802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.981893</td>\n",
       "      <td>2.710474</td>\n",
       "      <td>2.191160</td>\n",
       "      <td>0.789249</td>\n",
       "      <td>21.725583</td>\n",
       "      <td>33.685854</td>\n",
       "      <td>7.603477</td>\n",
       "      <td>11.981230</td>\n",
       "      <td>16.448616</td>\n",
       "      <td>1.010288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054343</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.961211</td>\n",
       "      <td>21.972497</td>\n",
       "      <td>0.976782</td>\n",
       "      <td>1.113990</td>\n",
       "      <td>4.810449</td>\n",
       "      <td>1.236143</td>\n",
       "      <td>0.928271</td>\n",
       "      <td>1.706164</td>\n",
       "      <td>3.073721</td>\n",
       "      <td>...</td>\n",
       "      <td>0.952303</td>\n",
       "      <td>4.391199</td>\n",
       "      <td>2.251523</td>\n",
       "      <td>0.831166</td>\n",
       "      <td>24.281447</td>\n",
       "      <td>56.189855</td>\n",
       "      <td>9.829170</td>\n",
       "      <td>8.417731</td>\n",
       "      <td>16.942906</td>\n",
       "      <td>0.879999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054344</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.802412</td>\n",
       "      <td>22.618258</td>\n",
       "      <td>1.146317</td>\n",
       "      <td>1.113299</td>\n",
       "      <td>4.048876</td>\n",
       "      <td>1.340815</td>\n",
       "      <td>0.880647</td>\n",
       "      <td>2.355578</td>\n",
       "      <td>3.902926</td>\n",
       "      <td>...</td>\n",
       "      <td>0.888738</td>\n",
       "      <td>3.320270</td>\n",
       "      <td>2.060882</td>\n",
       "      <td>0.811653</td>\n",
       "      <td>15.325906</td>\n",
       "      <td>58.047052</td>\n",
       "      <td>6.854160</td>\n",
       "      <td>7.689744</td>\n",
       "      <td>25.439824</td>\n",
       "      <td>1.039451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054345</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.964033</td>\n",
       "      <td>22.302836</td>\n",
       "      <td>1.128793</td>\n",
       "      <td>1.135271</td>\n",
       "      <td>2.373974</td>\n",
       "      <td>1.257381</td>\n",
       "      <td>0.874247</td>\n",
       "      <td>2.393149</td>\n",
       "      <td>6.094806</td>\n",
       "      <td>...</td>\n",
       "      <td>0.886234</td>\n",
       "      <td>3.197014</td>\n",
       "      <td>2.586177</td>\n",
       "      <td>0.835287</td>\n",
       "      <td>17.117737</td>\n",
       "      <td>58.464519</td>\n",
       "      <td>9.044523</td>\n",
       "      <td>6.061908</td>\n",
       "      <td>13.607890</td>\n",
       "      <td>1.062880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054346</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.097807</td>\n",
       "      <td>34.703123</td>\n",
       "      <td>0.950305</td>\n",
       "      <td>1.320155</td>\n",
       "      <td>1.274872</td>\n",
       "      <td>1.259774</td>\n",
       "      <td>0.881936</td>\n",
       "      <td>2.074347</td>\n",
       "      <td>2.414738</td>\n",
       "      <td>...</td>\n",
       "      <td>0.935212</td>\n",
       "      <td>2.186442</td>\n",
       "      <td>1.854861</td>\n",
       "      <td>1.056514</td>\n",
       "      <td>20.310387</td>\n",
       "      <td>60.301068</td>\n",
       "      <td>7.138543</td>\n",
       "      <td>14.245139</td>\n",
       "      <td>31.558448</td>\n",
       "      <td>0.901901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows × 13858 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Kidney Chromophobe      A1BG        A2M     A2ML1   A3GALT2  \\\n",
       "GSM1054230                 0.0  0.903276  27.154642  0.893226  1.357179   \n",
       "GSM1054231                 0.0  0.821580  17.070914  0.963602  1.423179   \n",
       "GSM1054232                 0.0  0.865428  34.490890  0.996714  1.518340   \n",
       "GSM1054233                 0.0  0.890772   9.773921  1.108153  1.352150   \n",
       "GSM1054234                 0.0  0.986434  31.282731  0.855068  1.415984   \n",
       "...                        ...       ...        ...       ...       ...   \n",
       "GSM1054342                 0.0  0.868972  22.165116  1.102940  1.279548   \n",
       "GSM1054343                 0.0  0.961211  21.972497  0.976782  1.113990   \n",
       "GSM1054344                 0.0  0.802412  22.618258  1.146317  1.113299   \n",
       "GSM1054345                 0.0  0.964033  22.302836  1.128793  1.135271   \n",
       "GSM1054346                 0.0  1.097807  34.703123  0.950305  1.320155   \n",
       "\n",
       "               A4GALT     A4GNT      AAA1      AAAS      AACS  ...      ZXDA  \\\n",
       "GSM1054230   2.008587  1.090213  1.113465  3.079174  4.593716  ...  0.956936   \n",
       "GSM1054231   1.581049  1.108882  0.962892  1.792671  8.201332  ...  0.911934   \n",
       "GSM1054232   2.094153  1.133496  1.238555  3.890664  4.255689  ...  0.868369   \n",
       "GSM1054233   1.300390  1.041011  0.957591  1.664235  4.674827  ...  0.892851   \n",
       "GSM1054234  11.355988  1.010757  0.926479  2.094559  2.133348  ...  0.953112   \n",
       "...               ...       ...       ...       ...       ...  ...       ...   \n",
       "GSM1054342   1.733522  1.089880  0.843426  1.531277  7.013802  ...  0.981893   \n",
       "GSM1054343   4.810449  1.236143  0.928271  1.706164  3.073721  ...  0.952303   \n",
       "GSM1054344   4.048876  1.340815  0.880647  2.355578  3.902926  ...  0.888738   \n",
       "GSM1054345   2.373974  1.257381  0.874247  2.393149  6.094806  ...  0.886234   \n",
       "GSM1054346   1.274872  1.259774  0.881936  2.074347  2.414738  ...  0.935212   \n",
       "\n",
       "                ZXDB      ZXDC    ZYG11A     ZYG11B        ZYX     ZZEF1  \\\n",
       "GSM1054230  5.579101  2.600057  1.052827  15.959863  19.655399  7.226102   \n",
       "GSM1054231  6.643073  1.840664  0.810640  53.815354  15.911316  9.249937   \n",
       "GSM1054232  4.749234  2.945843  1.003140  15.712620  18.813045  7.148919   \n",
       "GSM1054233  3.495437  1.402277  0.902783  31.327785   9.908477  6.130382   \n",
       "GSM1054234  3.221408  2.098021  1.003471  24.561863  37.106860  9.664492   \n",
       "...              ...       ...       ...        ...        ...       ...   \n",
       "GSM1054342  2.710474  2.191160  0.789249  21.725583  33.685854  7.603477   \n",
       "GSM1054343  4.391199  2.251523  0.831166  24.281447  56.189855  9.829170   \n",
       "GSM1054344  3.320270  2.060882  0.811653  15.325906  58.047052  6.854160   \n",
       "GSM1054345  3.197014  2.586177  0.835287  17.117737  58.464519  9.044523   \n",
       "GSM1054346  2.186442  1.854861  1.056514  20.310387  60.301068  7.138543   \n",
       "\n",
       "                 ZZZ3      EIF2A     RAB1C  \n",
       "GSM1054230  13.926936  36.250133  0.815878  \n",
       "GSM1054231   8.321252  21.393597  0.698088  \n",
       "GSM1054232  10.700753  24.619330  0.820355  \n",
       "GSM1054233   4.641731  10.641300  0.759291  \n",
       "GSM1054234  16.992292  28.841662  0.779781  \n",
       "...               ...        ...       ...  \n",
       "GSM1054342  11.981230  16.448616  1.010288  \n",
       "GSM1054343   8.417731  16.942906  0.879999  \n",
       "GSM1054344   7.689744  25.439824  1.039451  \n",
       "GSM1054345   6.061908  13.607890  1.062880  \n",
       "GSM1054346  14.245139  31.558448  0.901901  \n",
       "\n",
       "[117 rows x 13858 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 117 samples.\n",
      "For the feature 'Kidney Chromophobe', the least common label is '1.0' with 1 occurrences. This represents 0.85% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"An antioxidant response phenotype is shared between hereditary and sporadic type 2 papillary renal cell carcinoma\"\n",
      "!Series_summary\t\"Fumarate hydratase (FH) mutation causes hereditary type 2 papillary renal cell carcinoma (HLRCC, Hereditary Leiomyomatosis and Renal Cell Cancer (MM ID # 605839)). The main effect of FH mutation is fumarate accumulation. The current paradigm posits that the main consequence of fumarate accumulation is HIF-a stabilization. Paradoxically, FH mutation differs from other HIF-a stabilizing mutations, such as VHL and SDH mutations, in its associated tumor types. We identified that fumarate can directly up-regulate antioxidant response element (ARE)-controlled genes. We demonstrated that AKR1B10 is an ARE-controlled gene and is up-regulated upon FH knockdown as well as in FH-null cell lines. AKR1B10 overexpression is also a prominent feature in both hereditary and sporadic PRCC2. This phenotype better explains the similarities between hereditary and sporadic PRCC2.\"\n",
      "!Series_overall_design\t\"Expression profiling renal normal and tumor tissue\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM655513</th>\n",
       "      <th>GSM655514</th>\n",
       "      <th>GSM655515</th>\n",
       "      <th>GSM655516</th>\n",
       "      <th>GSM655517</th>\n",
       "      <th>GSM655518</th>\n",
       "      <th>GSM655519</th>\n",
       "      <th>GSM655520</th>\n",
       "      <th>GSM655521</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM655570</th>\n",
       "      <th>GSM655571</th>\n",
       "      <th>GSM655572</th>\n",
       "      <th>GSM655573</th>\n",
       "      <th>GSM655574</th>\n",
       "      <th>GSM655575</th>\n",
       "      <th>GSM655576</th>\n",
       "      <th>GSM655577</th>\n",
       "      <th>GSM655578</th>\n",
       "      <th>GSM655579</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: normal_tissue_from_ccRCC_patient</td>\n",
       "      <td>disease state: ccRCC</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: Pap_type2</td>\n",
       "      <td>disease state: Pap_type2</td>\n",
       "      <td>disease state: Pap_type2</td>\n",
       "      <td>disease state: HLRCC</td>\n",
       "      <td>disease state: HLRCC</td>\n",
       "      <td>disease state: HLRCC</td>\n",
       "      <td>disease state: HLRCC</td>\n",
       "      <td>disease state: HLRCC</td>\n",
       "      <td>disease state: normal_tissue_from_FH_patient</td>\n",
       "      <td>disease state: normal_tissue_from_FH_patient</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "\n",
       "                                         GSM655513  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655514  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655515  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655516  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655517  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655518  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655519  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient   \n",
       "\n",
       "                                         GSM655520             GSM655521  ...  \\\n",
       "0  disease state: normal_tissue_from_ccRCC_patient  disease state: ccRCC  ...   \n",
       "\n",
       "                  GSM655570                 GSM655571  \\\n",
       "0  disease state: Pap_type2  disease state: Pap_type2   \n",
       "\n",
       "                  GSM655572             GSM655573             GSM655574  \\\n",
       "0  disease state: Pap_type2  disease state: HLRCC  disease state: HLRCC   \n",
       "\n",
       "              GSM655575             GSM655576             GSM655577  \\\n",
       "0  disease state: HLRCC  disease state: HLRCC  disease state: HLRCC   \n",
       "\n",
       "                                      GSM655578  \\\n",
       "0  disease state: normal_tissue_from_FH_patient   \n",
       "\n",
       "                                      GSM655579  \n",
       "0  disease state: normal_tissue_from_FH_patient  \n",
       "\n",
       "[1 rows x 68 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished \n",
    "cohort = accession_num = \"GSE26574\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'disease state: normal_tissue_from_ccRCC_patient',\n",
       "       'disease state: ccRCC', 'disease state: Chromophobe',\n",
       "       'disease state: Pap_type1', 'disease state: Pap_type2',\n",
       "       'disease state: HLRCC',\n",
       "       'disease state: normal_tissue_from_FH_patient'], dtype=object)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease state: Chromophobe':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM655513</th>\n",
       "      <th>GSM655514</th>\n",
       "      <th>GSM655515</th>\n",
       "      <th>GSM655516</th>\n",
       "      <th>GSM655517</th>\n",
       "      <th>GSM655518</th>\n",
       "      <th>GSM655519</th>\n",
       "      <th>GSM655520</th>\n",
       "      <th>GSM655521</th>\n",
       "      <th>GSM655522</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM655570</th>\n",
       "      <th>GSM655571</th>\n",
       "      <th>GSM655572</th>\n",
       "      <th>GSM655573</th>\n",
       "      <th>GSM655574</th>\n",
       "      <th>GSM655575</th>\n",
       "      <th>GSM655576</th>\n",
       "      <th>GSM655577</th>\n",
       "      <th>GSM655578</th>\n",
       "      <th>GSM655579</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM655513  GSM655514  GSM655515  GSM655516  GSM655517  \\\n",
       "Kidney Chromophobe          0          0          0          0          0   \n",
       "\n",
       "                    GSM655518  GSM655519  GSM655520  GSM655521  GSM655522  \\\n",
       "Kidney Chromophobe          0          0          0          0          0   \n",
       "\n",
       "                    ...  GSM655570  GSM655571  GSM655572  GSM655573  \\\n",
       "Kidney Chromophobe  ...          0          0          0          0   \n",
       "\n",
       "                    GSM655574  GSM655575  GSM655576  GSM655577  GSM655578  \\\n",
       "Kidney Chromophobe          0          0          0          0          0   \n",
       "\n",
       "                    GSM655579  \n",
       "Kidney Chromophobe          0  \n",
       "\n",
       "[1 rows x 67 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM655513</th>\n",
       "      <th>GSM655514</th>\n",
       "      <th>GSM655515</th>\n",
       "      <th>GSM655516</th>\n",
       "      <th>GSM655517</th>\n",
       "      <th>GSM655518</th>\n",
       "      <th>GSM655519</th>\n",
       "      <th>GSM655520</th>\n",
       "      <th>GSM655521</th>\n",
       "      <th>GSM655522</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM655570</th>\n",
       "      <th>GSM655571</th>\n",
       "      <th>GSM655572</th>\n",
       "      <th>GSM655573</th>\n",
       "      <th>GSM655574</th>\n",
       "      <th>GSM655575</th>\n",
       "      <th>GSM655576</th>\n",
       "      <th>GSM655577</th>\n",
       "      <th>GSM655578</th>\n",
       "      <th>GSM655579</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5.668639</td>\n",
       "      <td>5.397250</td>\n",
       "      <td>5.533062</td>\n",
       "      <td>5.439878</td>\n",
       "      <td>5.455800</td>\n",
       "      <td>5.429943</td>\n",
       "      <td>5.433182</td>\n",
       "      <td>5.920859</td>\n",
       "      <td>5.403412</td>\n",
       "      <td>5.565034</td>\n",
       "      <td>...</td>\n",
       "      <td>6.223809</td>\n",
       "      <td>5.480571</td>\n",
       "      <td>6.147314</td>\n",
       "      <td>6.249860</td>\n",
       "      <td>5.163622</td>\n",
       "      <td>5.785643</td>\n",
       "      <td>5.347343</td>\n",
       "      <td>5.160851</td>\n",
       "      <td>5.617737</td>\n",
       "      <td>5.595228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12.464613</td>\n",
       "      <td>12.821652</td>\n",
       "      <td>12.966227</td>\n",
       "      <td>12.520486</td>\n",
       "      <td>13.089543</td>\n",
       "      <td>13.331817</td>\n",
       "      <td>12.606140</td>\n",
       "      <td>12.834138</td>\n",
       "      <td>12.828438</td>\n",
       "      <td>12.980657</td>\n",
       "      <td>...</td>\n",
       "      <td>11.032362</td>\n",
       "      <td>11.496659</td>\n",
       "      <td>9.691689</td>\n",
       "      <td>11.886333</td>\n",
       "      <td>9.689367</td>\n",
       "      <td>12.391459</td>\n",
       "      <td>10.935704</td>\n",
       "      <td>5.521329</td>\n",
       "      <td>12.925633</td>\n",
       "      <td>12.379518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7.919827</td>\n",
       "      <td>7.605824</td>\n",
       "      <td>7.350612</td>\n",
       "      <td>6.686158</td>\n",
       "      <td>6.956433</td>\n",
       "      <td>7.230992</td>\n",
       "      <td>7.650826</td>\n",
       "      <td>7.131763</td>\n",
       "      <td>7.106921</td>\n",
       "      <td>7.396125</td>\n",
       "      <td>...</td>\n",
       "      <td>7.089215</td>\n",
       "      <td>6.427457</td>\n",
       "      <td>6.477982</td>\n",
       "      <td>7.520532</td>\n",
       "      <td>6.666302</td>\n",
       "      <td>8.042353</td>\n",
       "      <td>8.071641</td>\n",
       "      <td>7.995131</td>\n",
       "      <td>8.272262</td>\n",
       "      <td>9.176026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7.004454</td>\n",
       "      <td>7.640562</td>\n",
       "      <td>6.839965</td>\n",
       "      <td>7.179755</td>\n",
       "      <td>6.764172</td>\n",
       "      <td>7.095299</td>\n",
       "      <td>6.867889</td>\n",
       "      <td>6.434765</td>\n",
       "      <td>6.639619</td>\n",
       "      <td>6.506634</td>\n",
       "      <td>...</td>\n",
       "      <td>7.201593</td>\n",
       "      <td>7.707917</td>\n",
       "      <td>7.483945</td>\n",
       "      <td>5.785292</td>\n",
       "      <td>6.091871</td>\n",
       "      <td>5.636089</td>\n",
       "      <td>5.852559</td>\n",
       "      <td>5.879160</td>\n",
       "      <td>6.439763</td>\n",
       "      <td>7.493414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>8.897253</td>\n",
       "      <td>8.377712</td>\n",
       "      <td>12.165405</td>\n",
       "      <td>7.766924</td>\n",
       "      <td>10.101467</td>\n",
       "      <td>10.485897</td>\n",
       "      <td>8.890075</td>\n",
       "      <td>7.136956</td>\n",
       "      <td>7.642404</td>\n",
       "      <td>7.404853</td>\n",
       "      <td>...</td>\n",
       "      <td>9.228735</td>\n",
       "      <td>12.464738</td>\n",
       "      <td>9.500725</td>\n",
       "      <td>9.865214</td>\n",
       "      <td>9.082017</td>\n",
       "      <td>9.470265</td>\n",
       "      <td>8.840548</td>\n",
       "      <td>6.709212</td>\n",
       "      <td>10.095208</td>\n",
       "      <td>7.925754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730202</th>\n",
       "      <td>5.148511</td>\n",
       "      <td>5.253727</td>\n",
       "      <td>5.452233</td>\n",
       "      <td>5.347934</td>\n",
       "      <td>5.208044</td>\n",
       "      <td>5.377347</td>\n",
       "      <td>5.174600</td>\n",
       "      <td>5.093092</td>\n",
       "      <td>5.318383</td>\n",
       "      <td>5.183217</td>\n",
       "      <td>...</td>\n",
       "      <td>5.391429</td>\n",
       "      <td>5.373495</td>\n",
       "      <td>5.269745</td>\n",
       "      <td>5.462639</td>\n",
       "      <td>5.438494</td>\n",
       "      <td>5.183750</td>\n",
       "      <td>5.008280</td>\n",
       "      <td>5.471475</td>\n",
       "      <td>5.096284</td>\n",
       "      <td>5.147228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730236</th>\n",
       "      <td>5.727775</td>\n",
       "      <td>6.254866</td>\n",
       "      <td>6.060818</td>\n",
       "      <td>6.010149</td>\n",
       "      <td>5.910481</td>\n",
       "      <td>5.935546</td>\n",
       "      <td>5.788366</td>\n",
       "      <td>6.451025</td>\n",
       "      <td>5.519453</td>\n",
       "      <td>5.638241</td>\n",
       "      <td>...</td>\n",
       "      <td>6.160716</td>\n",
       "      <td>6.208490</td>\n",
       "      <td>6.142807</td>\n",
       "      <td>5.313290</td>\n",
       "      <td>5.270430</td>\n",
       "      <td>5.354565</td>\n",
       "      <td>5.255737</td>\n",
       "      <td>4.981816</td>\n",
       "      <td>5.380207</td>\n",
       "      <td>4.973487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730241</th>\n",
       "      <td>5.731454</td>\n",
       "      <td>5.074041</td>\n",
       "      <td>5.453450</td>\n",
       "      <td>5.466722</td>\n",
       "      <td>5.308568</td>\n",
       "      <td>5.155898</td>\n",
       "      <td>5.093306</td>\n",
       "      <td>5.153164</td>\n",
       "      <td>5.210335</td>\n",
       "      <td>5.156344</td>\n",
       "      <td>...</td>\n",
       "      <td>5.089060</td>\n",
       "      <td>5.277337</td>\n",
       "      <td>5.135769</td>\n",
       "      <td>5.644737</td>\n",
       "      <td>5.108351</td>\n",
       "      <td>5.047582</td>\n",
       "      <td>4.824020</td>\n",
       "      <td>5.129366</td>\n",
       "      <td>5.220636</td>\n",
       "      <td>5.371268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730249</th>\n",
       "      <td>6.085765</td>\n",
       "      <td>6.258911</td>\n",
       "      <td>6.260450</td>\n",
       "      <td>6.228555</td>\n",
       "      <td>5.961106</td>\n",
       "      <td>6.102829</td>\n",
       "      <td>6.139825</td>\n",
       "      <td>5.957617</td>\n",
       "      <td>6.112640</td>\n",
       "      <td>6.009320</td>\n",
       "      <td>...</td>\n",
       "      <td>6.222924</td>\n",
       "      <td>6.466923</td>\n",
       "      <td>6.312049</td>\n",
       "      <td>5.432837</td>\n",
       "      <td>5.689616</td>\n",
       "      <td>5.510738</td>\n",
       "      <td>5.775398</td>\n",
       "      <td>5.785600</td>\n",
       "      <td>5.828588</td>\n",
       "      <td>5.799976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730269</th>\n",
       "      <td>7.056239</td>\n",
       "      <td>7.118928</td>\n",
       "      <td>7.238508</td>\n",
       "      <td>6.993614</td>\n",
       "      <td>6.897037</td>\n",
       "      <td>7.051942</td>\n",
       "      <td>7.020473</td>\n",
       "      <td>7.280863</td>\n",
       "      <td>6.792522</td>\n",
       "      <td>6.752757</td>\n",
       "      <td>...</td>\n",
       "      <td>7.328011</td>\n",
       "      <td>7.279786</td>\n",
       "      <td>7.015374</td>\n",
       "      <td>6.908914</td>\n",
       "      <td>7.013859</td>\n",
       "      <td>6.616630</td>\n",
       "      <td>6.939240</td>\n",
       "      <td>6.816868</td>\n",
       "      <td>6.408270</td>\n",
       "      <td>6.920472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17637 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM655513  GSM655514  GSM655515  GSM655516  GSM655517  GSM655518  \\\n",
       "ID                                                                         \n",
       "1        5.668639   5.397250   5.533062   5.439878   5.455800   5.429943   \n",
       "2       12.464613  12.821652  12.966227  12.520486  13.089543  13.331817   \n",
       "9        7.919827   7.605824   7.350612   6.686158   6.956433   7.230992   \n",
       "10       7.004454   7.640562   6.839965   7.179755   6.764172   7.095299   \n",
       "12       8.897253   8.377712  12.165405   7.766924  10.101467  10.485897   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "730202   5.148511   5.253727   5.452233   5.347934   5.208044   5.377347   \n",
       "730236   5.727775   6.254866   6.060818   6.010149   5.910481   5.935546   \n",
       "730241   5.731454   5.074041   5.453450   5.466722   5.308568   5.155898   \n",
       "730249   6.085765   6.258911   6.260450   6.228555   5.961106   6.102829   \n",
       "730269   7.056239   7.118928   7.238508   6.993614   6.897037   7.051942   \n",
       "\n",
       "        GSM655519  GSM655520  GSM655521  GSM655522  ...  GSM655570  GSM655571  \\\n",
       "ID                                                  ...                         \n",
       "1        5.433182   5.920859   5.403412   5.565034  ...   6.223809   5.480571   \n",
       "2       12.606140  12.834138  12.828438  12.980657  ...  11.032362  11.496659   \n",
       "9        7.650826   7.131763   7.106921   7.396125  ...   7.089215   6.427457   \n",
       "10       6.867889   6.434765   6.639619   6.506634  ...   7.201593   7.707917   \n",
       "12       8.890075   7.136956   7.642404   7.404853  ...   9.228735  12.464738   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "730202   5.174600   5.093092   5.318383   5.183217  ...   5.391429   5.373495   \n",
       "730236   5.788366   6.451025   5.519453   5.638241  ...   6.160716   6.208490   \n",
       "730241   5.093306   5.153164   5.210335   5.156344  ...   5.089060   5.277337   \n",
       "730249   6.139825   5.957617   6.112640   6.009320  ...   6.222924   6.466923   \n",
       "730269   7.020473   7.280863   6.792522   6.752757  ...   7.328011   7.279786   \n",
       "\n",
       "        GSM655572  GSM655573  GSM655574  GSM655575  GSM655576  GSM655577  \\\n",
       "ID                                                                         \n",
       "1        6.147314   6.249860   5.163622   5.785643   5.347343   5.160851   \n",
       "2        9.691689  11.886333   9.689367  12.391459  10.935704   5.521329   \n",
       "9        6.477982   7.520532   6.666302   8.042353   8.071641   7.995131   \n",
       "10       7.483945   5.785292   6.091871   5.636089   5.852559   5.879160   \n",
       "12       9.500725   9.865214   9.082017   9.470265   8.840548   6.709212   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "730202   5.269745   5.462639   5.438494   5.183750   5.008280   5.471475   \n",
       "730236   6.142807   5.313290   5.270430   5.354565   5.255737   4.981816   \n",
       "730241   5.135769   5.644737   5.108351   5.047582   4.824020   5.129366   \n",
       "730249   6.312049   5.432837   5.689616   5.510738   5.775398   5.785600   \n",
       "730269   7.015374   6.908914   7.013859   6.616630   6.939240   6.816868   \n",
       "\n",
       "        GSM655578  GSM655579  \n",
       "ID                            \n",
       "1        5.617737   5.595228  \n",
       "2       12.925633  12.379518  \n",
       "9        8.272262   9.176026  \n",
       "10       6.439763   7.493414  \n",
       "12      10.095208   7.925754  \n",
       "...           ...        ...  \n",
       "730202   5.096284   5.147228  \n",
       "730236   5.380207   4.973487  \n",
       "730241   5.220636   5.371268  \n",
       "730249   5.828588   5.799976  \n",
       "730269   6.408270   6.920472  \n",
       "\n",
       "[17637 rows x 67 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1', '10', '100', '1000', '10000'], 'CHR': ['19', '8', '20', '18', '1'], 'ORF': ['A1BG', 'NAT2', 'ADA', 'CDH2', 'AKT3'], 'GENE_ID': [1.0, 10.0, 100.0, 1000.0, 10000.0]}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1', '2', '9', '10', '12', '13', '14', '15', '16', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1', '10', '100', '1000', '10000'], 'CHR': ['19', '8', '20', '18', '1'], 'ORF': ['A1BG', 'NAT2', 'ADA', 'CDH2', 'AKT3'], 'GENE_ID': [1.0, 10.0, 100.0, 1000.0, 10000.0]}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'CHR', 'ORF', 'GENE_ID'], dtype='object')"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'ORF'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM655513</th>\n",
       "      <th>GSM655514</th>\n",
       "      <th>GSM655515</th>\n",
       "      <th>GSM655516</th>\n",
       "      <th>GSM655517</th>\n",
       "      <th>GSM655518</th>\n",
       "      <th>GSM655519</th>\n",
       "      <th>GSM655520</th>\n",
       "      <th>GSM655521</th>\n",
       "      <th>GSM655522</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM655570</th>\n",
       "      <th>GSM655571</th>\n",
       "      <th>GSM655572</th>\n",
       "      <th>GSM655573</th>\n",
       "      <th>GSM655574</th>\n",
       "      <th>GSM655575</th>\n",
       "      <th>GSM655576</th>\n",
       "      <th>GSM655577</th>\n",
       "      <th>GSM655578</th>\n",
       "      <th>GSM655579</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>5.668639</td>\n",
       "      <td>5.397250</td>\n",
       "      <td>5.533062</td>\n",
       "      <td>5.439878</td>\n",
       "      <td>5.455800</td>\n",
       "      <td>5.429943</td>\n",
       "      <td>5.433182</td>\n",
       "      <td>5.920859</td>\n",
       "      <td>5.403412</td>\n",
       "      <td>5.565034</td>\n",
       "      <td>...</td>\n",
       "      <td>6.223809</td>\n",
       "      <td>5.480571</td>\n",
       "      <td>6.147314</td>\n",
       "      <td>6.249860</td>\n",
       "      <td>5.163622</td>\n",
       "      <td>5.785643</td>\n",
       "      <td>5.347343</td>\n",
       "      <td>5.160851</td>\n",
       "      <td>5.617737</td>\n",
       "      <td>5.595228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>9.682636</td>\n",
       "      <td>8.724913</td>\n",
       "      <td>8.581043</td>\n",
       "      <td>6.410286</td>\n",
       "      <td>6.235908</td>\n",
       "      <td>7.171356</td>\n",
       "      <td>9.123482</td>\n",
       "      <td>8.831986</td>\n",
       "      <td>9.576187</td>\n",
       "      <td>8.656312</td>\n",
       "      <td>...</td>\n",
       "      <td>6.592295</td>\n",
       "      <td>7.518565</td>\n",
       "      <td>6.904563</td>\n",
       "      <td>5.550474</td>\n",
       "      <td>5.611425</td>\n",
       "      <td>5.478196</td>\n",
       "      <td>5.221965</td>\n",
       "      <td>5.535751</td>\n",
       "      <td>8.179168</td>\n",
       "      <td>9.224674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>12.464613</td>\n",
       "      <td>12.821652</td>\n",
       "      <td>12.966227</td>\n",
       "      <td>12.520486</td>\n",
       "      <td>13.089543</td>\n",
       "      <td>13.331817</td>\n",
       "      <td>12.606140</td>\n",
       "      <td>12.834138</td>\n",
       "      <td>12.828438</td>\n",
       "      <td>12.980657</td>\n",
       "      <td>...</td>\n",
       "      <td>11.032362</td>\n",
       "      <td>11.496659</td>\n",
       "      <td>9.691689</td>\n",
       "      <td>11.886333</td>\n",
       "      <td>9.689367</td>\n",
       "      <td>12.391459</td>\n",
       "      <td>10.935704</td>\n",
       "      <td>5.521329</td>\n",
       "      <td>12.925633</td>\n",
       "      <td>12.379518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>5.100448</td>\n",
       "      <td>5.234235</td>\n",
       "      <td>5.262083</td>\n",
       "      <td>5.108046</td>\n",
       "      <td>5.153058</td>\n",
       "      <td>5.005294</td>\n",
       "      <td>5.216862</td>\n",
       "      <td>4.988229</td>\n",
       "      <td>5.101942</td>\n",
       "      <td>5.133923</td>\n",
       "      <td>...</td>\n",
       "      <td>5.315465</td>\n",
       "      <td>5.292718</td>\n",
       "      <td>5.260594</td>\n",
       "      <td>5.120014</td>\n",
       "      <td>5.101100</td>\n",
       "      <td>5.063032</td>\n",
       "      <td>4.969653</td>\n",
       "      <td>5.054539</td>\n",
       "      <td>5.067441</td>\n",
       "      <td>5.026277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>8.787488</td>\n",
       "      <td>7.887596</td>\n",
       "      <td>8.429632</td>\n",
       "      <td>8.313462</td>\n",
       "      <td>8.144748</td>\n",
       "      <td>8.193744</td>\n",
       "      <td>8.415532</td>\n",
       "      <td>8.161773</td>\n",
       "      <td>7.909114</td>\n",
       "      <td>7.967972</td>\n",
       "      <td>...</td>\n",
       "      <td>7.563385</td>\n",
       "      <td>8.597225</td>\n",
       "      <td>7.648621</td>\n",
       "      <td>7.797198</td>\n",
       "      <td>7.435960</td>\n",
       "      <td>8.229121</td>\n",
       "      <td>7.069481</td>\n",
       "      <td>7.245178</td>\n",
       "      <td>8.511372</td>\n",
       "      <td>7.888782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>8.726466</td>\n",
       "      <td>8.622260</td>\n",
       "      <td>8.639624</td>\n",
       "      <td>8.487453</td>\n",
       "      <td>8.528002</td>\n",
       "      <td>8.808438</td>\n",
       "      <td>8.563243</td>\n",
       "      <td>8.759431</td>\n",
       "      <td>8.124979</td>\n",
       "      <td>8.527813</td>\n",
       "      <td>...</td>\n",
       "      <td>7.936377</td>\n",
       "      <td>8.133096</td>\n",
       "      <td>7.969016</td>\n",
       "      <td>8.531433</td>\n",
       "      <td>8.737743</td>\n",
       "      <td>8.471879</td>\n",
       "      <td>7.752223</td>\n",
       "      <td>7.706087</td>\n",
       "      <td>8.967396</td>\n",
       "      <td>8.762011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>8.857001</td>\n",
       "      <td>8.208841</td>\n",
       "      <td>8.259732</td>\n",
       "      <td>7.955382</td>\n",
       "      <td>8.092397</td>\n",
       "      <td>8.097794</td>\n",
       "      <td>8.159850</td>\n",
       "      <td>8.714595</td>\n",
       "      <td>8.282024</td>\n",
       "      <td>7.934380</td>\n",
       "      <td>...</td>\n",
       "      <td>7.860002</td>\n",
       "      <td>7.948557</td>\n",
       "      <td>7.781597</td>\n",
       "      <td>7.917889</td>\n",
       "      <td>8.242090</td>\n",
       "      <td>7.291694</td>\n",
       "      <td>7.943692</td>\n",
       "      <td>8.781280</td>\n",
       "      <td>8.781819</td>\n",
       "      <td>8.907930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>8.965583</td>\n",
       "      <td>9.547013</td>\n",
       "      <td>9.823586</td>\n",
       "      <td>9.714980</td>\n",
       "      <td>10.179625</td>\n",
       "      <td>10.434708</td>\n",
       "      <td>9.511976</td>\n",
       "      <td>8.886803</td>\n",
       "      <td>9.161385</td>\n",
       "      <td>9.508151</td>\n",
       "      <td>...</td>\n",
       "      <td>9.701222</td>\n",
       "      <td>10.041946</td>\n",
       "      <td>10.060402</td>\n",
       "      <td>9.926511</td>\n",
       "      <td>9.515641</td>\n",
       "      <td>10.052259</td>\n",
       "      <td>9.102406</td>\n",
       "      <td>9.584249</td>\n",
       "      <td>9.194147</td>\n",
       "      <td>8.390669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>7.983045</td>\n",
       "      <td>7.868412</td>\n",
       "      <td>8.190961</td>\n",
       "      <td>7.895259</td>\n",
       "      <td>7.864069</td>\n",
       "      <td>7.842018</td>\n",
       "      <td>7.840938</td>\n",
       "      <td>7.647074</td>\n",
       "      <td>7.597856</td>\n",
       "      <td>8.150198</td>\n",
       "      <td>...</td>\n",
       "      <td>8.143651</td>\n",
       "      <td>8.045217</td>\n",
       "      <td>7.738805</td>\n",
       "      <td>8.205857</td>\n",
       "      <td>8.328447</td>\n",
       "      <td>8.374914</td>\n",
       "      <td>8.196029</td>\n",
       "      <td>7.934793</td>\n",
       "      <td>8.317318</td>\n",
       "      <td>8.205114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>9.034118</td>\n",
       "      <td>8.738175</td>\n",
       "      <td>8.905796</td>\n",
       "      <td>8.933261</td>\n",
       "      <td>9.227768</td>\n",
       "      <td>8.879539</td>\n",
       "      <td>8.801854</td>\n",
       "      <td>8.985928</td>\n",
       "      <td>9.009432</td>\n",
       "      <td>9.335553</td>\n",
       "      <td>...</td>\n",
       "      <td>8.873733</td>\n",
       "      <td>8.867270</td>\n",
       "      <td>8.830455</td>\n",
       "      <td>9.798580</td>\n",
       "      <td>10.391157</td>\n",
       "      <td>10.172715</td>\n",
       "      <td>9.482569</td>\n",
       "      <td>10.650001</td>\n",
       "      <td>9.670159</td>\n",
       "      <td>9.573885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15043 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM655513  GSM655514  GSM655515  GSM655516  GSM655517  GSM655518  \\\n",
       "A1BG     5.668639   5.397250   5.533062   5.439878   5.455800   5.429943   \n",
       "A1CF     9.682636   8.724913   8.581043   6.410286   6.235908   7.171356   \n",
       "A2M     12.464613  12.821652  12.966227  12.520486  13.089543  13.331817   \n",
       "A2ML1    5.100448   5.234235   5.262083   5.108046   5.153058   5.005294   \n",
       "A4GALT   8.787488   7.887596   8.429632   8.313462   8.144748   8.193744   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZXDC     8.726466   8.622260   8.639624   8.487453   8.528002   8.808438   \n",
       "ZYG11B   8.857001   8.208841   8.259732   7.955382   8.092397   8.097794   \n",
       "ZYX      8.965583   9.547013   9.823586   9.714980  10.179625  10.434708   \n",
       "ZZEF1    7.983045   7.868412   8.190961   7.895259   7.864069   7.842018   \n",
       "ZZZ3     9.034118   8.738175   8.905796   8.933261   9.227768   8.879539   \n",
       "\n",
       "        GSM655519  GSM655520  GSM655521  GSM655522  ...  GSM655570  GSM655571  \\\n",
       "A1BG     5.433182   5.920859   5.403412   5.565034  ...   6.223809   5.480571   \n",
       "A1CF     9.123482   8.831986   9.576187   8.656312  ...   6.592295   7.518565   \n",
       "A2M     12.606140  12.834138  12.828438  12.980657  ...  11.032362  11.496659   \n",
       "A2ML1    5.216862   4.988229   5.101942   5.133923  ...   5.315465   5.292718   \n",
       "A4GALT   8.415532   8.161773   7.909114   7.967972  ...   7.563385   8.597225   \n",
       "...           ...        ...        ...        ...  ...        ...        ...   \n",
       "ZXDC     8.563243   8.759431   8.124979   8.527813  ...   7.936377   8.133096   \n",
       "ZYG11B   8.159850   8.714595   8.282024   7.934380  ...   7.860002   7.948557   \n",
       "ZYX      9.511976   8.886803   9.161385   9.508151  ...   9.701222  10.041946   \n",
       "ZZEF1    7.840938   7.647074   7.597856   8.150198  ...   8.143651   8.045217   \n",
       "ZZZ3     8.801854   8.985928   9.009432   9.335553  ...   8.873733   8.867270   \n",
       "\n",
       "        GSM655572  GSM655573  GSM655574  GSM655575  GSM655576  GSM655577  \\\n",
       "A1BG     6.147314   6.249860   5.163622   5.785643   5.347343   5.160851   \n",
       "A1CF     6.904563   5.550474   5.611425   5.478196   5.221965   5.535751   \n",
       "A2M      9.691689  11.886333   9.689367  12.391459  10.935704   5.521329   \n",
       "A2ML1    5.260594   5.120014   5.101100   5.063032   4.969653   5.054539   \n",
       "A4GALT   7.648621   7.797198   7.435960   8.229121   7.069481   7.245178   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "ZXDC     7.969016   8.531433   8.737743   8.471879   7.752223   7.706087   \n",
       "ZYG11B   7.781597   7.917889   8.242090   7.291694   7.943692   8.781280   \n",
       "ZYX     10.060402   9.926511   9.515641  10.052259   9.102406   9.584249   \n",
       "ZZEF1    7.738805   8.205857   8.328447   8.374914   8.196029   7.934793   \n",
       "ZZZ3     8.830455   9.798580  10.391157  10.172715   9.482569  10.650001   \n",
       "\n",
       "        GSM655578  GSM655579  \n",
       "A1BG     5.617737   5.595228  \n",
       "A1CF     8.179168   9.224674  \n",
       "A2M     12.925633  12.379518  \n",
       "A2ML1    5.067441   5.026277  \n",
       "A4GALT   8.511372   7.888782  \n",
       "...           ...        ...  \n",
       "ZXDC     8.967396   8.762011  \n",
       "ZYG11B   8.781819   8.907930  \n",
       "ZYX      9.194147   8.390669  \n",
       "ZZEF1    8.317318   8.205114  \n",
       "ZZZ3     9.670159   9.573885  \n",
       "\n",
       "[15043 rows x 67 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM655513</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.668639</td>\n",
       "      <td>9.682636</td>\n",
       "      <td>12.464613</td>\n",
       "      <td>5.100448</td>\n",
       "      <td>8.787488</td>\n",
       "      <td>5.809985</td>\n",
       "      <td>5.371351</td>\n",
       "      <td>8.812667</td>\n",
       "      <td>8.677541</td>\n",
       "      <td>...</td>\n",
       "      <td>8.344515</td>\n",
       "      <td>6.125898</td>\n",
       "      <td>8.179752</td>\n",
       "      <td>5.803459</td>\n",
       "      <td>5.454540</td>\n",
       "      <td>8.726466</td>\n",
       "      <td>8.857001</td>\n",
       "      <td>8.965583</td>\n",
       "      <td>7.983045</td>\n",
       "      <td>9.034118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655514</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.397250</td>\n",
       "      <td>8.724913</td>\n",
       "      <td>12.821652</td>\n",
       "      <td>5.234235</td>\n",
       "      <td>7.887596</td>\n",
       "      <td>5.884389</td>\n",
       "      <td>5.552994</td>\n",
       "      <td>8.726981</td>\n",
       "      <td>8.543634</td>\n",
       "      <td>...</td>\n",
       "      <td>8.441628</td>\n",
       "      <td>6.190711</td>\n",
       "      <td>7.928924</td>\n",
       "      <td>5.754128</td>\n",
       "      <td>5.740305</td>\n",
       "      <td>8.622260</td>\n",
       "      <td>8.208841</td>\n",
       "      <td>9.547013</td>\n",
       "      <td>7.868412</td>\n",
       "      <td>8.738175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655515</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.533062</td>\n",
       "      <td>8.581043</td>\n",
       "      <td>12.966227</td>\n",
       "      <td>5.262083</td>\n",
       "      <td>8.429632</td>\n",
       "      <td>5.849380</td>\n",
       "      <td>5.497786</td>\n",
       "      <td>9.041170</td>\n",
       "      <td>8.688982</td>\n",
       "      <td>...</td>\n",
       "      <td>8.252270</td>\n",
       "      <td>5.488916</td>\n",
       "      <td>8.042933</td>\n",
       "      <td>5.734307</td>\n",
       "      <td>5.589368</td>\n",
       "      <td>8.639624</td>\n",
       "      <td>8.259732</td>\n",
       "      <td>9.823586</td>\n",
       "      <td>8.190961</td>\n",
       "      <td>8.905796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655516</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.439878</td>\n",
       "      <td>6.410286</td>\n",
       "      <td>12.520486</td>\n",
       "      <td>5.108046</td>\n",
       "      <td>8.313462</td>\n",
       "      <td>5.709369</td>\n",
       "      <td>5.475360</td>\n",
       "      <td>9.044497</td>\n",
       "      <td>9.346827</td>\n",
       "      <td>...</td>\n",
       "      <td>8.370992</td>\n",
       "      <td>6.272480</td>\n",
       "      <td>9.006933</td>\n",
       "      <td>5.655809</td>\n",
       "      <td>5.579518</td>\n",
       "      <td>8.487453</td>\n",
       "      <td>7.955382</td>\n",
       "      <td>9.714980</td>\n",
       "      <td>7.895259</td>\n",
       "      <td>8.933261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655517</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.455800</td>\n",
       "      <td>6.235908</td>\n",
       "      <td>13.089543</td>\n",
       "      <td>5.153058</td>\n",
       "      <td>8.144748</td>\n",
       "      <td>5.623220</td>\n",
       "      <td>5.497168</td>\n",
       "      <td>8.881100</td>\n",
       "      <td>8.633818</td>\n",
       "      <td>...</td>\n",
       "      <td>8.230474</td>\n",
       "      <td>6.430454</td>\n",
       "      <td>8.604294</td>\n",
       "      <td>5.667027</td>\n",
       "      <td>5.701254</td>\n",
       "      <td>8.528002</td>\n",
       "      <td>8.092397</td>\n",
       "      <td>10.179625</td>\n",
       "      <td>7.864069</td>\n",
       "      <td>9.227768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655575</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.785643</td>\n",
       "      <td>5.478196</td>\n",
       "      <td>12.391459</td>\n",
       "      <td>5.063032</td>\n",
       "      <td>8.229121</td>\n",
       "      <td>6.011650</td>\n",
       "      <td>4.933610</td>\n",
       "      <td>8.040029</td>\n",
       "      <td>10.093184</td>\n",
       "      <td>...</td>\n",
       "      <td>8.840957</td>\n",
       "      <td>6.061439</td>\n",
       "      <td>10.494962</td>\n",
       "      <td>5.642196</td>\n",
       "      <td>5.644396</td>\n",
       "      <td>8.471879</td>\n",
       "      <td>7.291694</td>\n",
       "      <td>10.052259</td>\n",
       "      <td>8.374914</td>\n",
       "      <td>10.172715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655576</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.347343</td>\n",
       "      <td>5.221965</td>\n",
       "      <td>10.935704</td>\n",
       "      <td>4.969653</td>\n",
       "      <td>7.069481</td>\n",
       "      <td>5.365046</td>\n",
       "      <td>4.906179</td>\n",
       "      <td>8.122486</td>\n",
       "      <td>8.354842</td>\n",
       "      <td>...</td>\n",
       "      <td>9.702138</td>\n",
       "      <td>7.870020</td>\n",
       "      <td>11.444870</td>\n",
       "      <td>5.998233</td>\n",
       "      <td>5.740564</td>\n",
       "      <td>7.752223</td>\n",
       "      <td>7.943692</td>\n",
       "      <td>9.102406</td>\n",
       "      <td>8.196029</td>\n",
       "      <td>9.482569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655577</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.160851</td>\n",
       "      <td>5.535751</td>\n",
       "      <td>5.521329</td>\n",
       "      <td>5.054539</td>\n",
       "      <td>7.245178</td>\n",
       "      <td>5.697122</td>\n",
       "      <td>5.170637</td>\n",
       "      <td>8.477883</td>\n",
       "      <td>9.645716</td>\n",
       "      <td>...</td>\n",
       "      <td>9.061160</td>\n",
       "      <td>9.209461</td>\n",
       "      <td>11.673874</td>\n",
       "      <td>5.224771</td>\n",
       "      <td>6.182081</td>\n",
       "      <td>7.706087</td>\n",
       "      <td>8.781280</td>\n",
       "      <td>9.584249</td>\n",
       "      <td>7.934793</td>\n",
       "      <td>10.650001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655578</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.617737</td>\n",
       "      <td>8.179168</td>\n",
       "      <td>12.925633</td>\n",
       "      <td>5.067441</td>\n",
       "      <td>8.511372</td>\n",
       "      <td>5.065305</td>\n",
       "      <td>5.023733</td>\n",
       "      <td>8.062029</td>\n",
       "      <td>8.667378</td>\n",
       "      <td>...</td>\n",
       "      <td>8.833000</td>\n",
       "      <td>5.645277</td>\n",
       "      <td>8.825598</td>\n",
       "      <td>5.962845</td>\n",
       "      <td>5.752671</td>\n",
       "      <td>8.967396</td>\n",
       "      <td>8.781819</td>\n",
       "      <td>9.194147</td>\n",
       "      <td>8.317318</td>\n",
       "      <td>9.670159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM655579</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.595228</td>\n",
       "      <td>9.224674</td>\n",
       "      <td>12.379518</td>\n",
       "      <td>5.026277</td>\n",
       "      <td>7.888782</td>\n",
       "      <td>5.501903</td>\n",
       "      <td>5.247685</td>\n",
       "      <td>8.101225</td>\n",
       "      <td>8.506567</td>\n",
       "      <td>...</td>\n",
       "      <td>8.656634</td>\n",
       "      <td>5.803004</td>\n",
       "      <td>8.783044</td>\n",
       "      <td>5.844714</td>\n",
       "      <td>5.584350</td>\n",
       "      <td>8.762011</td>\n",
       "      <td>8.907930</td>\n",
       "      <td>8.390669</td>\n",
       "      <td>8.205114</td>\n",
       "      <td>9.573885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 15044 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Kidney Chromophobe      A1BG      A1CF        A2M     A2ML1  \\\n",
       "GSM655513                 0.0  5.668639  9.682636  12.464613  5.100448   \n",
       "GSM655514                 0.0  5.397250  8.724913  12.821652  5.234235   \n",
       "GSM655515                 0.0  5.533062  8.581043  12.966227  5.262083   \n",
       "GSM655516                 0.0  5.439878  6.410286  12.520486  5.108046   \n",
       "GSM655517                 0.0  5.455800  6.235908  13.089543  5.153058   \n",
       "...                       ...       ...       ...        ...       ...   \n",
       "GSM655575                 0.0  5.785643  5.478196  12.391459  5.063032   \n",
       "GSM655576                 0.0  5.347343  5.221965  10.935704  4.969653   \n",
       "GSM655577                 0.0  5.160851  5.535751   5.521329  5.054539   \n",
       "GSM655578                 0.0  5.617737  8.179168  12.925633  5.067441   \n",
       "GSM655579                 0.0  5.595228  9.224674  12.379518  5.026277   \n",
       "\n",
       "             A4GALT     A4GNT      AAA1      AAAS       AACS  ...      ZW10  \\\n",
       "GSM655513  8.787488  5.809985  5.371351  8.812667   8.677541  ...  8.344515   \n",
       "GSM655514  7.887596  5.884389  5.552994  8.726981   8.543634  ...  8.441628   \n",
       "GSM655515  8.429632  5.849380  5.497786  9.041170   8.688982  ...  8.252270   \n",
       "GSM655516  8.313462  5.709369  5.475360  9.044497   9.346827  ...  8.370992   \n",
       "GSM655517  8.144748  5.623220  5.497168  8.881100   8.633818  ...  8.230474   \n",
       "...             ...       ...       ...       ...        ...  ...       ...   \n",
       "GSM655575  8.229121  6.011650  4.933610  8.040029  10.093184  ...  8.840957   \n",
       "GSM655576  7.069481  5.365046  4.906179  8.122486   8.354842  ...  9.702138   \n",
       "GSM655577  7.245178  5.697122  5.170637  8.477883   9.645716  ...  9.061160   \n",
       "GSM655578  8.511372  5.065305  5.023733  8.062029   8.667378  ...  8.833000   \n",
       "GSM655579  7.888782  5.501903  5.247685  8.101225   8.506567  ...  8.656634   \n",
       "\n",
       "             ZWILCH      ZWINT      ZXDA      ZXDB      ZXDC    ZYG11B  \\\n",
       "GSM655513  6.125898   8.179752  5.803459  5.454540  8.726466  8.857001   \n",
       "GSM655514  6.190711   7.928924  5.754128  5.740305  8.622260  8.208841   \n",
       "GSM655515  5.488916   8.042933  5.734307  5.589368  8.639624  8.259732   \n",
       "GSM655516  6.272480   9.006933  5.655809  5.579518  8.487453  7.955382   \n",
       "GSM655517  6.430454   8.604294  5.667027  5.701254  8.528002  8.092397   \n",
       "...             ...        ...       ...       ...       ...       ...   \n",
       "GSM655575  6.061439  10.494962  5.642196  5.644396  8.471879  7.291694   \n",
       "GSM655576  7.870020  11.444870  5.998233  5.740564  7.752223  7.943692   \n",
       "GSM655577  9.209461  11.673874  5.224771  6.182081  7.706087  8.781280   \n",
       "GSM655578  5.645277   8.825598  5.962845  5.752671  8.967396  8.781819   \n",
       "GSM655579  5.803004   8.783044  5.844714  5.584350  8.762011  8.907930   \n",
       "\n",
       "                 ZYX     ZZEF1       ZZZ3  \n",
       "GSM655513   8.965583  7.983045   9.034118  \n",
       "GSM655514   9.547013  7.868412   8.738175  \n",
       "GSM655515   9.823586  8.190961   8.905796  \n",
       "GSM655516   9.714980  7.895259   8.933261  \n",
       "GSM655517  10.179625  7.864069   9.227768  \n",
       "...              ...       ...        ...  \n",
       "GSM655575  10.052259  8.374914  10.172715  \n",
       "GSM655576   9.102406  8.196029   9.482569  \n",
       "GSM655577   9.584249  7.934793  10.650001  \n",
       "GSM655578   9.194147  8.317318   9.670159  \n",
       "GSM655579   8.390669  8.205114   9.573885  \n",
       "\n",
       "[67 rows x 15044 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 67 samples.\n",
      "For the feature 'Kidney Chromophobe', the least common label is '1.0' with 10 occurrences. This represents 14.93% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Virtual-Karyotyping with SNP microarrays in morphologically challenging renal cell neoplasms\"\n",
      "!Series_summary\t\"Genetic lesions characteristic for RCC subtypes can be identified by virtual karyotyping with SNP microarrays. In this study, we examined whether virtual karyotypes could be used to better classify a cohort of morphologically challenging/unclassified RCC.\"\n",
      "!Series_overall_design\t\"Tumor resection specimens from 17 patients were profiled by virtual karyotyping with Affymetrix 10K 2.0 or 250K Nsp SNP Mapping arrays and were also evaluated independently by a panel of seven genito-urinary pathologists. Tumors were classified by the established pattern of genomic imbalances based on a reference cohort of 98 cases with classic morphology and compared to the morphologic diagnosis of the pathologist panel. In 3 cases, samples from areas with different morphologic appearance were also tested (n=5).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM240385</th>\n",
       "      <th>GSM240386</th>\n",
       "      <th>GSM240387</th>\n",
       "      <th>GSM240388</th>\n",
       "      <th>GSM240389</th>\n",
       "      <th>GSM240390</th>\n",
       "      <th>GSM240391</th>\n",
       "      <th>GSM240392</th>\n",
       "      <th>GSM240393</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM366129</th>\n",
       "      <th>GSM366130</th>\n",
       "      <th>GSM366131</th>\n",
       "      <th>GSM366132</th>\n",
       "      <th>GSM366133</th>\n",
       "      <th>GSM366135</th>\n",
       "      <th>GSM366136</th>\n",
       "      <th>GSM366137</th>\n",
       "      <th>GSM366138</th>\n",
       "      <th>GSM366139</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: N/A</td>\n",
       "      <td>Sex: N/A</td>\n",
       "      <td>Sex: N/A</td>\n",
       "      <td>Sex: N/A</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: N/A</td>\n",
       "      <td>Sex: N/A</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>agedecade: 7</td>\n",
       "      <td>agedecade: 5</td>\n",
       "      <td>agedecade: 6</td>\n",
       "      <td>agedecade: 7</td>\n",
       "      <td>agedecade: 5</td>\n",
       "      <td>agedecade: 5</td>\n",
       "      <td>agedecade: 7</td>\n",
       "      <td>agedecade: 6</td>\n",
       "      <td>agedecade: 8</td>\n",
       "      <td>...</td>\n",
       "      <td>AgeDecade: 68</td>\n",
       "      <td>AgeDecade: N/A</td>\n",
       "      <td>AgeDecade: N/A</td>\n",
       "      <td>AgeDecade: N/A</td>\n",
       "      <td>AgeDecade: N/A</td>\n",
       "      <td>AgeDecade: 68</td>\n",
       "      <td>AgeDecade: N/A</td>\n",
       "      <td>AgeDecade: N/A</td>\n",
       "      <td>AgeDecade: 68</td>\n",
       "      <td>AgeDecade: 68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor size: 7.7</td>\n",
       "      <td>tumor size: 9.7</td>\n",
       "      <td>tumor size: 5.5</td>\n",
       "      <td>tumor size: 7</td>\n",
       "      <td>tumor size: 6</td>\n",
       "      <td>tumor size: 4</td>\n",
       "      <td>tumor size: 6</td>\n",
       "      <td>tumor size: 5.4</td>\n",
       "      <td>tumor size: 4.5</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor Size: 5.699999809</td>\n",
       "      <td>Tumor Size: N/A</td>\n",
       "      <td>Tumor Size: N/A</td>\n",
       "      <td>Tumor Size: N/A</td>\n",
       "      <td>Tumor Size: 6</td>\n",
       "      <td>Tumor Size: 5</td>\n",
       "      <td>Tumor Size: N/A</td>\n",
       "      <td>Tumor Size: N/A</td>\n",
       "      <td>Tumor Size: N/A</td>\n",
       "      <td>Tumor Size: 14.19999981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>furhman nuclear grade: 2</td>\n",
       "      <td>furhman nuclear grade: 3</td>\n",
       "      <td>furhman nuclear grade: 3</td>\n",
       "      <td>furhman nuclear grade: 3</td>\n",
       "      <td>furhman nuclear grade: 2</td>\n",
       "      <td>furhman nuclear grade: 2</td>\n",
       "      <td>furhman nuclear grade: 2</td>\n",
       "      <td>furhman nuclear grade: 2</td>\n",
       "      <td>furhman nuclear grade: 3</td>\n",
       "      <td>...</td>\n",
       "      <td>Furhman Nuclear Grade: 4</td>\n",
       "      <td>Furhman Nuclear Grade: N/A</td>\n",
       "      <td>Furhman Nuclear Grade: N/A</td>\n",
       "      <td>Furhman Nuclear Grade: N/A</td>\n",
       "      <td>Furhman Nuclear Grade: 4</td>\n",
       "      <td>Furhman Nuclear Grade: 2</td>\n",
       "      <td>Furhman Nuclear Grade: N/A</td>\n",
       "      <td>Furhman Nuclear Grade: N/A</td>\n",
       "      <td>Furhman Nuclear Grade: 3</td>\n",
       "      <td>Furhman Nuclear Grade: 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tnm stage: T2NXMX</td>\n",
       "      <td>tnm stage: T2NXMX</td>\n",
       "      <td>tnm stage: T1NXMX</td>\n",
       "      <td>tnm stage: T3bNXMX</td>\n",
       "      <td>tnm stage: T1bNXMX</td>\n",
       "      <td>tnm stage: T1aNXMX</td>\n",
       "      <td>tnm stage: T1bNXMX</td>\n",
       "      <td>tnm stage: T1bNXMX</td>\n",
       "      <td>tnm stage: T1bNXMX</td>\n",
       "      <td>...</td>\n",
       "      <td>TNM Stage: T3bNxMx</td>\n",
       "      <td>TNM Stage: N/A</td>\n",
       "      <td>TNM Stage: N/A</td>\n",
       "      <td>TNM Stage: N/A</td>\n",
       "      <td>TNM Stage: T1bNxMx</td>\n",
       "      <td>TNM Stage: T1NN0MX</td>\n",
       "      <td>TNM Stage: N/A</td>\n",
       "      <td>TNM Stage: N/A</td>\n",
       "      <td>TNM Stage: N/A</td>\n",
       "      <td>TNM Stage: T3aN0Mx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: Y</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>lymphovascular invasion: N</td>\n",
       "      <td>...</td>\n",
       "      <td>LymphoVascular Invasion: FALSE</td>\n",
       "      <td>LymphoVascular Invasion: N/A</td>\n",
       "      <td>LymphoVascular Invasion: N/A</td>\n",
       "      <td>LymphoVascular Invasion: N/A</td>\n",
       "      <td>LymphoVascular Invasion: FALSE</td>\n",
       "      <td>LymphoVascular Invasion: FALSE</td>\n",
       "      <td>LymphoVascular Invasion: N/A</td>\n",
       "      <td>LymphoVascular Invasion: N/A</td>\n",
       "      <td>LymphoVascular Invasion: FALSE</td>\n",
       "      <td>LymphoVascular Invasion: FALSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gross appearance: WC solid, fleshy tan-black</td>\n",
       "      <td>gross appearance: Yellow orange granular lesio...</td>\n",
       "      <td>gross appearance: WC yellow variegated</td>\n",
       "      <td>gross appearance: Tan-brown with central scar ...</td>\n",
       "      <td>gross appearance: Bright yellow</td>\n",
       "      <td>gross appearance: lobulated yellow</td>\n",
       "      <td>gross appearance: variegated golden yellow</td>\n",
       "      <td>gross appearance: golden yellow orange</td>\n",
       "      <td>gross appearance: variegated golden yellow</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>comment: IHC:  pos CK7, EP4, CI; neg vim, CD10...</td>\n",
       "      <td>comment:</td>\n",
       "      <td>comment:</td>\n",
       "      <td>comment: .015 cm PRCC</td>\n",
       "      <td>comment:</td>\n",
       "      <td>comment: Multifocal; IHC:  pos RCC; neg CK7, E...</td>\n",
       "      <td>comment:</td>\n",
       "      <td>comment:</td>\n",
       "      <td>comment:</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "5  !Sample_characteristics_ch1   \n",
       "6  !Sample_characteristics_ch1   \n",
       "7  !Sample_characteristics_ch1   \n",
       "\n",
       "                                           GSM240385  \\\n",
       "0                                             sex: F   \n",
       "1                                       agedecade: 7   \n",
       "2                                    tumor size: 7.7   \n",
       "3                           furhman nuclear grade: 2   \n",
       "4                                  tnm stage: T2NXMX   \n",
       "5                         lymphovascular invasion: N   \n",
       "6       gross appearance: WC solid, fleshy tan-black   \n",
       "7  comment: IHC:  pos CK7, EP4, CI; neg vim, CD10...   \n",
       "\n",
       "                                           GSM240386  \\\n",
       "0                                             sex: M   \n",
       "1                                       agedecade: 5   \n",
       "2                                    tumor size: 9.7   \n",
       "3                           furhman nuclear grade: 3   \n",
       "4                                  tnm stage: T2NXMX   \n",
       "5                         lymphovascular invasion: N   \n",
       "6  gross appearance: Yellow orange granular lesio...   \n",
       "7                                          comment:    \n",
       "\n",
       "                                GSM240387  \\\n",
       "0                                  sex: M   \n",
       "1                            agedecade: 6   \n",
       "2                         tumor size: 5.5   \n",
       "3                furhman nuclear grade: 3   \n",
       "4                       tnm stage: T1NXMX   \n",
       "5              lymphovascular invasion: N   \n",
       "6  gross appearance: WC yellow variegated   \n",
       "7                               comment:    \n",
       "\n",
       "                                           GSM240388  \\\n",
       "0                                             sex: M   \n",
       "1                                       agedecade: 7   \n",
       "2                                      tumor size: 7   \n",
       "3                           furhman nuclear grade: 3   \n",
       "4                                 tnm stage: T3bNXMX   \n",
       "5                         lymphovascular invasion: Y   \n",
       "6  gross appearance: Tan-brown with central scar ...   \n",
       "7                              comment: .015 cm PRCC   \n",
       "\n",
       "                         GSM240389  \\\n",
       "0                           sex: F   \n",
       "1                     agedecade: 5   \n",
       "2                    tumor size: 6   \n",
       "3         furhman nuclear grade: 2   \n",
       "4               tnm stage: T1bNXMX   \n",
       "5       lymphovascular invasion: N   \n",
       "6  gross appearance: Bright yellow   \n",
       "7                        comment:    \n",
       "\n",
       "                                           GSM240390  \\\n",
       "0                                             sex: M   \n",
       "1                                       agedecade: 5   \n",
       "2                                      tumor size: 4   \n",
       "3                           furhman nuclear grade: 2   \n",
       "4                                 tnm stage: T1aNXMX   \n",
       "5                         lymphovascular invasion: N   \n",
       "6                 gross appearance: lobulated yellow   \n",
       "7  comment: Multifocal; IHC:  pos RCC; neg CK7, E...   \n",
       "\n",
       "                                    GSM240391  \\\n",
       "0                                      sex: M   \n",
       "1                                agedecade: 7   \n",
       "2                               tumor size: 6   \n",
       "3                    furhman nuclear grade: 2   \n",
       "4                          tnm stage: T1bNXMX   \n",
       "5                  lymphovascular invasion: N   \n",
       "6  gross appearance: variegated golden yellow   \n",
       "7                                   comment:    \n",
       "\n",
       "                                GSM240392  \\\n",
       "0                                  sex: F   \n",
       "1                            agedecade: 6   \n",
       "2                         tumor size: 5.4   \n",
       "3                furhman nuclear grade: 2   \n",
       "4                      tnm stage: T1bNXMX   \n",
       "5              lymphovascular invasion: N   \n",
       "6  gross appearance: golden yellow orange   \n",
       "7                               comment:    \n",
       "\n",
       "                                    GSM240393  ...  \\\n",
       "0                                      sex: M  ...   \n",
       "1                                agedecade: 8  ...   \n",
       "2                             tumor size: 4.5  ...   \n",
       "3                    furhman nuclear grade: 3  ...   \n",
       "4                          tnm stage: T1bNXMX  ...   \n",
       "5                  lymphovascular invasion: N  ...   \n",
       "6  gross appearance: variegated golden yellow  ...   \n",
       "7                                   comment:   ...   \n",
       "\n",
       "                        GSM366129                     GSM366130  \\\n",
       "0                          Sex: M                      Sex: N/A   \n",
       "1                   AgeDecade: 68                AgeDecade: N/A   \n",
       "2         Tumor Size: 5.699999809               Tumor Size: N/A   \n",
       "3        Furhman Nuclear Grade: 4    Furhman Nuclear Grade: N/A   \n",
       "4              TNM Stage: T3bNxMx                TNM Stage: N/A   \n",
       "5  LymphoVascular Invasion: FALSE  LymphoVascular Invasion: N/A   \n",
       "6                             NaN                           NaN   \n",
       "7                             NaN                           NaN   \n",
       "\n",
       "                      GSM366131                     GSM366132  \\\n",
       "0                      Sex: N/A                      Sex: N/A   \n",
       "1                AgeDecade: N/A                AgeDecade: N/A   \n",
       "2               Tumor Size: N/A               Tumor Size: N/A   \n",
       "3    Furhman Nuclear Grade: N/A    Furhman Nuclear Grade: N/A   \n",
       "4                TNM Stage: N/A                TNM Stage: N/A   \n",
       "5  LymphoVascular Invasion: N/A  LymphoVascular Invasion: N/A   \n",
       "6                           NaN                           NaN   \n",
       "7                           NaN                           NaN   \n",
       "\n",
       "                        GSM366133                       GSM366135  \\\n",
       "0                        Sex: N/A                          Sex: M   \n",
       "1                  AgeDecade: N/A                   AgeDecade: 68   \n",
       "2                   Tumor Size: 6                   Tumor Size: 5   \n",
       "3        Furhman Nuclear Grade: 4        Furhman Nuclear Grade: 2   \n",
       "4              TNM Stage: T1bNxMx              TNM Stage: T1NN0MX   \n",
       "5  LymphoVascular Invasion: FALSE  LymphoVascular Invasion: FALSE   \n",
       "6                             NaN                             NaN   \n",
       "7                             NaN                             NaN   \n",
       "\n",
       "                      GSM366136                     GSM366137  \\\n",
       "0                      Sex: N/A                      Sex: N/A   \n",
       "1                AgeDecade: N/A                AgeDecade: N/A   \n",
       "2               Tumor Size: N/A               Tumor Size: N/A   \n",
       "3    Furhman Nuclear Grade: N/A    Furhman Nuclear Grade: N/A   \n",
       "4                TNM Stage: N/A                TNM Stage: N/A   \n",
       "5  LymphoVascular Invasion: N/A  LymphoVascular Invasion: N/A   \n",
       "6                           NaN                           NaN   \n",
       "7                           NaN                           NaN   \n",
       "\n",
       "                        GSM366138                       GSM366139  \n",
       "0                          Sex: M                          Sex: F  \n",
       "1                   AgeDecade: 68                   AgeDecade: 68  \n",
       "2                 Tumor Size: N/A         Tumor Size: 14.19999981  \n",
       "3        Furhman Nuclear Grade: 3        Furhman Nuclear Grade: 4  \n",
       "4                  TNM Stage: N/A              TNM Stage: T3aN0Mx  \n",
       "5  LymphoVascular Invasion: FALSE  LymphoVascular Invasion: FALSE  \n",
       "6                             NaN                             NaN  \n",
       "7                             NaN                             NaN  \n",
       "\n",
       "[8 rows x 92 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No obvious trait convert\n",
    "cohort = accession_num = \"GSE14670\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'comment: IHC:  pos CK7, EP4, CI; neg vim, CD10, ecad, inh; Ki67 pos < 10%',\n",
       "       'comment: ', 'comment: .015 cm PRCC',\n",
       "       'comment: Multifocal; IHC:  pos RCC; neg CK7, EP4, vim, inh, IC; CD10 pos memb; ecad focal weak memb',\n",
       "       'comment: IHC:  RCC neg, CK7 scat pos and tubules.',\n",
       "       'comment: Multifocal; no IHC; no FISH',\n",
       "       'comment: 4cm OC, no ECE, no LVI',\n",
       "       'comment: IHC:  pos CD10, ecad, inh; neg CI, vim, rcc, EP4; CK7 scat pos; Ki67 1+ of neoplastic cells',\n",
       "       'comment: 21cm OC, no ECE, no LVI; FISH not c/w OC.',\n",
       "       'comment: Small focus of clear cell; multifocal vs tangential cut.',\n",
       "       'comment: Extensive necrosis.', nan], dtype=object)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[7]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'condition: tumor':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Expression analysis and in silico characterization of intronic long noncoding RNAs in renal cell carcinoma: emerging functional associations (RCC malignancy)\"\n",
      "!Series_summary\t\"Intronic and intergenic long noncoding RNAs (lncRNAs) are emerging gene expression regulators. The molecular pathogenesis of renal cell carcinoma (RCC) is still poorly understood, and in particular, limited studies are available for intronic lncRNAs expressed in RCC. Microarray experiments were performed with two different custom-designed arrays enriched with probes for lncRNAs mapping to intronic genomic regions. Samples from 18 primary clear cell RCC tumors and 11 nontumor adjacent matched tissues were analyzed with 4k-probes microarrays. Oligoarrays with 44k-probes were used to interrogate 17 RCC samples (14 clear cell, 2 papillary, 1 chromophobe subtypes) split into four pools. Meta-analyses were performed by taking the genomic coordinates of the RCC-expressed lncRNAs, and cross-referencing them with microarray expression data from three additional human tissues (normal liver, prostate tumor and kidney nontumor samples), and with large-scale public data for epigenetic regulatory marks and for evolutionarily conserved sequences. A signature of 29 intronic lncRNAs differentially expressed between RCC and nontumor samples was obtained (false discovery rate (FDR) <5%). An additional signature of 26 intronic lncRNAs significantly correlated with the RCC five-year patient survival outcome was identified (FDR <5%, p-value ≤0.01). We identified 4303 intronic antisense lncRNAs expressed in RCC, of which 25% were cis correlated (r >|0.6|) with the expression of the mRNA in the same locus across three human tissues. Gene Ontology (GO) analysis of those loci pointed to ‘regulation of biological processes’ as the main enriched category. A module map analysis of all expressed protein-coding genes in RCC that had a significant (r ≥|0.8|) trans correlation with the 20% most abundant lncRNAs identified 35 relevant (p <0.05) GO sets. In addition, we determined that 60% of these lncRNAs are evolutionarily conserved. At the genomic loci containing the intronic RCC-expressed lncRNAs, a strong association (p <0.001) was found between their transcription start sites and genomic marks such as CpG islands and histones methylation and acetylation. Intronic antisense lncRNAs are widely expressed in RCC tumors. Some of them are significantly altered in RCC in comparison with nontumor samples. The majority of these lncRNAs is evolutionarily conserved and possibly modulated by epigenetic modifications. Our data suggest that these RCC lncRNAs may contribute to the complex network of regulatory RNAs playing a role in renal cell malignant transformation.\"\n",
      "!Series_overall_design\t\"A total of 22 human kidney tissue samples consisting of 11 primary renal tumors and 11 matched adjacent nontumor tissues from clear cell renal cell carcinmoa (RCC) patients were evaluated in this study. We compared the expression profiles of tumor and non-tumor samples obtained from patients with clear cell RCC to evaluate a possible correlation of the lncRNAs with renal malignancy. The set of clear cell RCC expression profiles was generated using a custom-designed cDNA microarray platform with 4,608 unique elements in replicate (9,216) enriched in gene fragments that map to intronic regions of known human genes (GPL3985).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1004655</th>\n",
       "      <th>GSM1004656</th>\n",
       "      <th>GSM1004657</th>\n",
       "      <th>GSM1004658</th>\n",
       "      <th>GSM1004659</th>\n",
       "      <th>GSM1004660</th>\n",
       "      <th>GSM1004661</th>\n",
       "      <th>GSM1004662</th>\n",
       "      <th>GSM1004663</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1004689</th>\n",
       "      <th>GSM1004690</th>\n",
       "      <th>GSM1004691</th>\n",
       "      <th>GSM1004692</th>\n",
       "      <th>GSM1004693</th>\n",
       "      <th>GSM1004694</th>\n",
       "      <th>GSM1004695</th>\n",
       "      <th>GSM1004696</th>\n",
       "      <th>GSM1004697</th>\n",
       "      <th>GSM1004698</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>patient identifier: 3</td>\n",
       "      <td>patient identifier: 3</td>\n",
       "      <td>patient identifier: 5</td>\n",
       "      <td>patient identifier: 5</td>\n",
       "      <td>patient identifier: 8</td>\n",
       "      <td>patient identifier: 8</td>\n",
       "      <td>patient identifier: 9</td>\n",
       "      <td>patient identifier: 9</td>\n",
       "      <td>patient identifier: 10</td>\n",
       "      <td>...</td>\n",
       "      <td>patient identifier: 24</td>\n",
       "      <td>patient identifier: 24</td>\n",
       "      <td>patient identifier: 26</td>\n",
       "      <td>patient identifier: 26</td>\n",
       "      <td>patient identifier: 28</td>\n",
       "      <td>patient identifier: 28</td>\n",
       "      <td>patient identifier: 30</td>\n",
       "      <td>patient identifier: 30</td>\n",
       "      <td>patient identifier: 31</td>\n",
       "      <td>patient identifier: 31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>...</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "      <td>disease: clear cell renal cell carcinoma (RCC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>tissue: adjacent nontumor kidney tissue</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "      <td>tissue: primary kidney tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age at surgery (yrs): 78</td>\n",
       "      <td>age at surgery (yrs): 78</td>\n",
       "      <td>age at surgery (yrs): 53</td>\n",
       "      <td>age at surgery (yrs): 53</td>\n",
       "      <td>age at surgery (yrs): 71</td>\n",
       "      <td>age at surgery (yrs): 71</td>\n",
       "      <td>age at surgery (yrs): 39</td>\n",
       "      <td>age at surgery (yrs): 39</td>\n",
       "      <td>age at surgery (yrs): 34</td>\n",
       "      <td>...</td>\n",
       "      <td>age at surgery (yrs): 75</td>\n",
       "      <td>age at surgery (yrs): 75</td>\n",
       "      <td>age at surgery (yrs): 40</td>\n",
       "      <td>age at surgery (yrs): 40</td>\n",
       "      <td>age at surgery (yrs): 51</td>\n",
       "      <td>age at surgery (yrs): 51</td>\n",
       "      <td>age at surgery (yrs): 51</td>\n",
       "      <td>age at surgery (yrs): 51</td>\n",
       "      <td>age at surgery (yrs): 50</td>\n",
       "      <td>age at surgery (yrs): 50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: dead from other causes</td>\n",
       "      <td>patient status: dead from other causes</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>...</td>\n",
       "      <td>fuhrman grade: IV</td>\n",
       "      <td>fuhrman grade: IV</td>\n",
       "      <td>fuhrman grade: II</td>\n",
       "      <td>fuhrman grade: II</td>\n",
       "      <td>fuhrman grade: III</td>\n",
       "      <td>fuhrman grade: III</td>\n",
       "      <td>fuhrman grade: II</td>\n",
       "      <td>fuhrman grade: II</td>\n",
       "      <td>fuhrman grade: III</td>\n",
       "      <td>fuhrman grade: III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor size (cm): 15</td>\n",
       "      <td>tumor size (cm): 15</td>\n",
       "      <td>tumor size (cm): 7</td>\n",
       "      <td>tumor size (cm): 7</td>\n",
       "      <td>tumor size (cm): 5</td>\n",
       "      <td>tumor size (cm): 5</td>\n",
       "      <td>tumor size (cm): 8.5</td>\n",
       "      <td>tumor size (cm): 8.5</td>\n",
       "      <td>tumor size (cm): 8</td>\n",
       "      <td>tumor size (cm): 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>necrosis: yes</td>\n",
       "      <td>necrosis: yes</td>\n",
       "      <td>necrosis: no</td>\n",
       "      <td>necrosis: no</td>\n",
       "      <td>necrosis: no</td>\n",
       "      <td>necrosis: no</td>\n",
       "      <td>necrosis: yes</td>\n",
       "      <td>necrosis: yes</td>\n",
       "      <td>necrosis: yes</td>\n",
       "      <td>necrosis: yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: no</td>\n",
       "      <td>capsule infiltration: no</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "      <td>capsule infiltration: yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>tnm classification (t): 2</td>\n",
       "      <td>tnm classification (t): 2</td>\n",
       "      <td>tnm classification (t): 3b</td>\n",
       "      <td>tnm classification (t): 3b</td>\n",
       "      <td>tnm classification (t): 1</td>\n",
       "      <td>tnm classification (t): 1</td>\n",
       "      <td>tnm classification (t): 2</td>\n",
       "      <td>tnm classification (t): 2</td>\n",
       "      <td>tnm classification (t): 3b</td>\n",
       "      <td>tnm classification (t): 3b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): no data available</td>\n",
       "      <td>tnm classification (n): 2</td>\n",
       "      <td>tnm classification (n): 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): no data available</td>\n",
       "      <td>tnm classification (m): 1</td>\n",
       "      <td>tnm classification (m): 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: no data available</td>\n",
       "      <td>organ metastasis at surgery: lung, adjacent ti...</td>\n",
       "      <td>organ metastasis at surgery: lung, adjacent ti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>organ metastasis after surgery: liver, spleen</td>\n",
       "      <td>organ metastasis after surgery: liver, spleen</td>\n",
       "      <td>organ metastasis after surgery: no data available</td>\n",
       "      <td>organ metastasis after surgery: no data available</td>\n",
       "      <td>organ metastasis after surgery: no data available</td>\n",
       "      <td>organ metastasis after surgery: no data available</td>\n",
       "      <td>organ metastasis after surgery: brain, lung, b...</td>\n",
       "      <td>organ metastasis after surgery: brain, lung, b...</td>\n",
       "      <td>organ metastasis after surgery: no data available</td>\n",
       "      <td>organ metastasis after surgery: no data available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>patient status: alive without cancer</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: cancer-specific death</td>\n",
       "      <td>patient status: alive with cancer</td>\n",
       "      <td>patient status: alive with cancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          !Sample_geo_accession  \\\n",
       "0   !Sample_characteristics_ch1   \n",
       "1   !Sample_characteristics_ch1   \n",
       "2   !Sample_characteristics_ch1   \n",
       "3   !Sample_characteristics_ch1   \n",
       "4   !Sample_characteristics_ch1   \n",
       "5   !Sample_characteristics_ch1   \n",
       "6   !Sample_characteristics_ch1   \n",
       "7   !Sample_characteristics_ch1   \n",
       "8   !Sample_characteristics_ch1   \n",
       "9   !Sample_characteristics_ch1   \n",
       "10  !Sample_characteristics_ch1   \n",
       "11  !Sample_characteristics_ch1   \n",
       "12  !Sample_characteristics_ch1   \n",
       "13  !Sample_characteristics_ch1   \n",
       "14  !Sample_characteristics_ch1   \n",
       "\n",
       "                                        GSM1004655  \\\n",
       "0                            patient identifier: 3   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                   gender: female   \n",
       "4                         age at surgery (yrs): 78   \n",
       "5            patient status: cancer-specific death   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004656  \\\n",
       "0                            patient identifier: 3   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                   gender: female   \n",
       "4                         age at surgery (yrs): 78   \n",
       "5            patient status: cancer-specific death   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004657  \\\n",
       "0                            patient identifier: 5   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                     gender: male   \n",
       "4                         age at surgery (yrs): 53   \n",
       "5            patient status: cancer-specific death   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004658  \\\n",
       "0                            patient identifier: 5   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                     gender: male   \n",
       "4                         age at surgery (yrs): 53   \n",
       "5            patient status: cancer-specific death   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004659  \\\n",
       "0                            patient identifier: 8   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                   gender: female   \n",
       "4                         age at surgery (yrs): 71   \n",
       "5           patient status: dead from other causes   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004660  \\\n",
       "0                            patient identifier: 8   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                   gender: female   \n",
       "4                         age at surgery (yrs): 71   \n",
       "5           patient status: dead from other causes   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004661  \\\n",
       "0                            patient identifier: 9   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                     gender: male   \n",
       "4                         age at surgery (yrs): 39   \n",
       "5             patient status: alive without cancer   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004662  \\\n",
       "0                            patient identifier: 9   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2          tissue: adjacent nontumor kidney tissue   \n",
       "3                                     gender: male   \n",
       "4                         age at surgery (yrs): 39   \n",
       "5             patient status: alive without cancer   \n",
       "6                                              NaN   \n",
       "7                                              NaN   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             NaN   \n",
       "\n",
       "                                        GSM1004663  ...  \\\n",
       "0                           patient identifier: 10  ...   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)  ...   \n",
       "2          tissue: adjacent nontumor kidney tissue  ...   \n",
       "3                                   gender: female  ...   \n",
       "4                         age at surgery (yrs): 34  ...   \n",
       "5             patient status: alive without cancer  ...   \n",
       "6                                              NaN  ...   \n",
       "7                                              NaN  ...   \n",
       "8                                              NaN  ...   \n",
       "9                                              NaN  ...   \n",
       "10                                             NaN  ...   \n",
       "11                                             NaN  ...   \n",
       "12                                             NaN  ...   \n",
       "13                                             NaN  ...   \n",
       "14                                             NaN  ...   \n",
       "\n",
       "                                        GSM1004689  \\\n",
       "0                           patient identifier: 24   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                     tissue: primary kidney tumor   \n",
       "3                                     gender: male   \n",
       "4                         age at surgery (yrs): 75   \n",
       "5                                fuhrman grade: IV   \n",
       "6                              tumor size (cm): 15   \n",
       "7                                    necrosis: yes   \n",
       "8                        capsule infiltration: yes   \n",
       "9                        tnm classification (t): 2   \n",
       "10       tnm classification (n): no data available   \n",
       "11       tnm classification (m): no data available   \n",
       "12  organ metastasis at surgery: no data available   \n",
       "13   organ metastasis after surgery: liver, spleen   \n",
       "14           patient status: cancer-specific death   \n",
       "\n",
       "                                        GSM1004690  \\\n",
       "0                           patient identifier: 24   \n",
       "1   disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                     tissue: primary kidney tumor   \n",
       "3                                     gender: male   \n",
       "4                         age at surgery (yrs): 75   \n",
       "5                                fuhrman grade: IV   \n",
       "6                              tumor size (cm): 15   \n",
       "7                                    necrosis: yes   \n",
       "8                        capsule infiltration: yes   \n",
       "9                        tnm classification (t): 2   \n",
       "10       tnm classification (n): no data available   \n",
       "11       tnm classification (m): no data available   \n",
       "12  organ metastasis at surgery: no data available   \n",
       "13   organ metastasis after surgery: liver, spleen   \n",
       "14           patient status: cancer-specific death   \n",
       "\n",
       "                                           GSM1004691  \\\n",
       "0                              patient identifier: 26   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                      gender: female   \n",
       "4                            age at surgery (yrs): 40   \n",
       "5                                   fuhrman grade: II   \n",
       "6                                  tumor size (cm): 7   \n",
       "7                                        necrosis: no   \n",
       "8                           capsule infiltration: yes   \n",
       "9                          tnm classification (t): 3b   \n",
       "10          tnm classification (n): no data available   \n",
       "11          tnm classification (m): no data available   \n",
       "12     organ metastasis at surgery: no data available   \n",
       "13  organ metastasis after surgery: no data available   \n",
       "14               patient status: alive without cancer   \n",
       "\n",
       "                                           GSM1004692  \\\n",
       "0                              patient identifier: 26   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                      gender: female   \n",
       "4                            age at surgery (yrs): 40   \n",
       "5                                   fuhrman grade: II   \n",
       "6                                  tumor size (cm): 7   \n",
       "7                                        necrosis: no   \n",
       "8                           capsule infiltration: yes   \n",
       "9                          tnm classification (t): 3b   \n",
       "10          tnm classification (n): no data available   \n",
       "11          tnm classification (m): no data available   \n",
       "12     organ metastasis at surgery: no data available   \n",
       "13  organ metastasis after surgery: no data available   \n",
       "14               patient status: alive without cancer   \n",
       "\n",
       "                                           GSM1004693  \\\n",
       "0                              patient identifier: 28   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                        gender: male   \n",
       "4                            age at surgery (yrs): 51   \n",
       "5                                  fuhrman grade: III   \n",
       "6                                  tumor size (cm): 5   \n",
       "7                                        necrosis: no   \n",
       "8                            capsule infiltration: no   \n",
       "9                           tnm classification (t): 1   \n",
       "10          tnm classification (n): no data available   \n",
       "11          tnm classification (m): no data available   \n",
       "12     organ metastasis at surgery: no data available   \n",
       "13  organ metastasis after surgery: no data available   \n",
       "14               patient status: alive without cancer   \n",
       "\n",
       "                                           GSM1004694  \\\n",
       "0                              patient identifier: 28   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                        gender: male   \n",
       "4                            age at surgery (yrs): 51   \n",
       "5                                  fuhrman grade: III   \n",
       "6                                  tumor size (cm): 5   \n",
       "7                                        necrosis: no   \n",
       "8                            capsule infiltration: no   \n",
       "9                           tnm classification (t): 1   \n",
       "10          tnm classification (n): no data available   \n",
       "11          tnm classification (m): no data available   \n",
       "12     organ metastasis at surgery: no data available   \n",
       "13  organ metastasis after surgery: no data available   \n",
       "14               patient status: alive without cancer   \n",
       "\n",
       "                                           GSM1004695  \\\n",
       "0                              patient identifier: 30   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                      gender: female   \n",
       "4                            age at surgery (yrs): 51   \n",
       "5                                   fuhrman grade: II   \n",
       "6                                tumor size (cm): 8.5   \n",
       "7                                       necrosis: yes   \n",
       "8                           capsule infiltration: yes   \n",
       "9                           tnm classification (t): 2   \n",
       "10          tnm classification (n): no data available   \n",
       "11          tnm classification (m): no data available   \n",
       "12     organ metastasis at surgery: no data available   \n",
       "13  organ metastasis after surgery: brain, lung, b...   \n",
       "14              patient status: cancer-specific death   \n",
       "\n",
       "                                           GSM1004696  \\\n",
       "0                              patient identifier: 30   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                      gender: female   \n",
       "4                            age at surgery (yrs): 51   \n",
       "5                                   fuhrman grade: II   \n",
       "6                                tumor size (cm): 8.5   \n",
       "7                                       necrosis: yes   \n",
       "8                           capsule infiltration: yes   \n",
       "9                           tnm classification (t): 2   \n",
       "10          tnm classification (n): no data available   \n",
       "11          tnm classification (m): no data available   \n",
       "12     organ metastasis at surgery: no data available   \n",
       "13  organ metastasis after surgery: brain, lung, b...   \n",
       "14              patient status: cancer-specific death   \n",
       "\n",
       "                                           GSM1004697  \\\n",
       "0                              patient identifier: 31   \n",
       "1      disease: clear cell renal cell carcinoma (RCC)   \n",
       "2                        tissue: primary kidney tumor   \n",
       "3                                        gender: male   \n",
       "4                            age at surgery (yrs): 50   \n",
       "5                                  fuhrman grade: III   \n",
       "6                                  tumor size (cm): 8   \n",
       "7                                       necrosis: yes   \n",
       "8                           capsule infiltration: yes   \n",
       "9                          tnm classification (t): 3b   \n",
       "10                          tnm classification (n): 2   \n",
       "11                          tnm classification (m): 1   \n",
       "12  organ metastasis at surgery: lung, adjacent ti...   \n",
       "13  organ metastasis after surgery: no data available   \n",
       "14                  patient status: alive with cancer   \n",
       "\n",
       "                                           GSM1004698  \n",
       "0                              patient identifier: 31  \n",
       "1      disease: clear cell renal cell carcinoma (RCC)  \n",
       "2                        tissue: primary kidney tumor  \n",
       "3                                        gender: male  \n",
       "4                            age at surgery (yrs): 50  \n",
       "5                                  fuhrman grade: III  \n",
       "6                                  tumor size (cm): 8  \n",
       "7                                       necrosis: yes  \n",
       "8                           capsule infiltration: yes  \n",
       "9                          tnm classification (t): 3b  \n",
       "10                          tnm classification (n): 2  \n",
       "11                          tnm classification (m): 1  \n",
       "12  organ metastasis at surgery: lung, adjacent ti...  \n",
       "13  organ metastasis after surgery: no data available  \n",
       "14                  patient status: alive with cancer  \n",
       "\n",
       "[15 rows x 45 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE40911\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "\n",
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'disease: clear cell renal cell carcinoma (RCC)'], dtype=object)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 4\n",
    "gender_row = 3\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'disease: clear cell renal cell carcinoma (RCC)':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1004655</th>\n",
       "      <th>GSM1004656</th>\n",
       "      <th>GSM1004657</th>\n",
       "      <th>GSM1004658</th>\n",
       "      <th>GSM1004659</th>\n",
       "      <th>GSM1004660</th>\n",
       "      <th>GSM1004661</th>\n",
       "      <th>GSM1004662</th>\n",
       "      <th>GSM1004663</th>\n",
       "      <th>GSM1004664</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1004689</th>\n",
       "      <th>GSM1004690</th>\n",
       "      <th>GSM1004691</th>\n",
       "      <th>GSM1004692</th>\n",
       "      <th>GSM1004693</th>\n",
       "      <th>GSM1004694</th>\n",
       "      <th>GSM1004695</th>\n",
       "      <th>GSM1004696</th>\n",
       "      <th>GSM1004697</th>\n",
       "      <th>GSM1004698</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM1004655  GSM1004656  GSM1004657  GSM1004658  \\\n",
       "Kidney Chromophobe           1           1           1           1   \n",
       "Age                         78          78          53          53   \n",
       "Gender                       1           1           0           0   \n",
       "\n",
       "                    GSM1004659  GSM1004660  GSM1004661  GSM1004662  \\\n",
       "Kidney Chromophobe           1           1           1           1   \n",
       "Age                         71          71          39          39   \n",
       "Gender                       1           1           0           0   \n",
       "\n",
       "                    GSM1004663  GSM1004664  ...  GSM1004689  GSM1004690  \\\n",
       "Kidney Chromophobe           1           1  ...           1           1   \n",
       "Age                         34          34  ...          75          75   \n",
       "Gender                       1           1  ...           0           0   \n",
       "\n",
       "                    GSM1004691  GSM1004692  GSM1004693  GSM1004694  \\\n",
       "Kidney Chromophobe           1           1           1           1   \n",
       "Age                         40          40          51          51   \n",
       "Gender                       1           1           0           0   \n",
       "\n",
       "                    GSM1004695  GSM1004696  GSM1004697  GSM1004698  \n",
       "Kidney Chromophobe           1           1           1           1  \n",
       "Age                         51          51          50          50  \n",
       "Gender                       1           1           0           0  \n",
       "\n",
       "[3 rows x 44 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1004655</th>\n",
       "      <th>GSM1004656</th>\n",
       "      <th>GSM1004657</th>\n",
       "      <th>GSM1004658</th>\n",
       "      <th>GSM1004659</th>\n",
       "      <th>GSM1004660</th>\n",
       "      <th>GSM1004661</th>\n",
       "      <th>GSM1004662</th>\n",
       "      <th>GSM1004663</th>\n",
       "      <th>GSM1004664</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1004689</th>\n",
       "      <th>GSM1004690</th>\n",
       "      <th>GSM1004691</th>\n",
       "      <th>GSM1004692</th>\n",
       "      <th>GSM1004693</th>\n",
       "      <th>GSM1004694</th>\n",
       "      <th>GSM1004695</th>\n",
       "      <th>GSM1004696</th>\n",
       "      <th>GSM1004697</th>\n",
       "      <th>GSM1004698</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.437273</td>\n",
       "      <td>1.674227</td>\n",
       "      <td>5.305341</td>\n",
       "      <td>4.193955</td>\n",
       "      <td>1.227295</td>\n",
       "      <td>3.416227</td>\n",
       "      <td>3.718273</td>\n",
       "      <td>4.329773</td>\n",
       "      <td>2.180568</td>\n",
       "      <td>1.933909</td>\n",
       "      <td>...</td>\n",
       "      <td>2.263227</td>\n",
       "      <td>1.810614</td>\n",
       "      <td>5.394136</td>\n",
       "      <td>5.185057</td>\n",
       "      <td>2.817341</td>\n",
       "      <td>2.631955</td>\n",
       "      <td>1.718864</td>\n",
       "      <td>0.911318</td>\n",
       "      <td>1.290841</td>\n",
       "      <td>1.184500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.049591</td>\n",
       "      <td>2.043773</td>\n",
       "      <td>4.909659</td>\n",
       "      <td>4.769750</td>\n",
       "      <td>1.303795</td>\n",
       "      <td>2.549125</td>\n",
       "      <td>5.019545</td>\n",
       "      <td>5.817341</td>\n",
       "      <td>2.505500</td>\n",
       "      <td>2.462614</td>\n",
       "      <td>...</td>\n",
       "      <td>2.949727</td>\n",
       "      <td>2.252273</td>\n",
       "      <td>8.628886</td>\n",
       "      <td>9.682500</td>\n",
       "      <td>3.649705</td>\n",
       "      <td>3.624045</td>\n",
       "      <td>0.911318</td>\n",
       "      <td>1.885591</td>\n",
       "      <td>3.894159</td>\n",
       "      <td>2.964523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6.138932</td>\n",
       "      <td>5.641727</td>\n",
       "      <td>11.108727</td>\n",
       "      <td>10.532295</td>\n",
       "      <td>4.600114</td>\n",
       "      <td>9.868205</td>\n",
       "      <td>10.034773</td>\n",
       "      <td>12.374955</td>\n",
       "      <td>6.749909</td>\n",
       "      <td>6.394932</td>\n",
       "      <td>...</td>\n",
       "      <td>9.937273</td>\n",
       "      <td>9.117500</td>\n",
       "      <td>12.210477</td>\n",
       "      <td>14.781477</td>\n",
       "      <td>10.952080</td>\n",
       "      <td>10.515000</td>\n",
       "      <td>10.150250</td>\n",
       "      <td>8.656409</td>\n",
       "      <td>5.062489</td>\n",
       "      <td>4.377955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.655091</td>\n",
       "      <td>1.115705</td>\n",
       "      <td>6.016773</td>\n",
       "      <td>6.079091</td>\n",
       "      <td>1.459250</td>\n",
       "      <td>2.676705</td>\n",
       "      <td>1.828693</td>\n",
       "      <td>2.787205</td>\n",
       "      <td>2.109114</td>\n",
       "      <td>2.007932</td>\n",
       "      <td>...</td>\n",
       "      <td>2.164375</td>\n",
       "      <td>1.674795</td>\n",
       "      <td>4.452432</td>\n",
       "      <td>5.394136</td>\n",
       "      <td>2.623705</td>\n",
       "      <td>2.234193</td>\n",
       "      <td>1.003591</td>\n",
       "      <td>1.744398</td>\n",
       "      <td>1.441341</td>\n",
       "      <td>1.398080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.207614</td>\n",
       "      <td>2.803170</td>\n",
       "      <td>4.166545</td>\n",
       "      <td>3.309591</td>\n",
       "      <td>0.425273</td>\n",
       "      <td>0.958455</td>\n",
       "      <td>1.397477</td>\n",
       "      <td>2.026455</td>\n",
       "      <td>1.268511</td>\n",
       "      <td>1.314568</td>\n",
       "      <td>...</td>\n",
       "      <td>1.655739</td>\n",
       "      <td>1.375727</td>\n",
       "      <td>3.354136</td>\n",
       "      <td>4.187477</td>\n",
       "      <td>2.204364</td>\n",
       "      <td>1.943034</td>\n",
       "      <td>1.316227</td>\n",
       "      <td>1.314568</td>\n",
       "      <td>0.830182</td>\n",
       "      <td>1.011409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4536</th>\n",
       "      <td>6.042318</td>\n",
       "      <td>5.310705</td>\n",
       "      <td>8.327068</td>\n",
       "      <td>8.044136</td>\n",
       "      <td>5.652591</td>\n",
       "      <td>5.437250</td>\n",
       "      <td>3.649705</td>\n",
       "      <td>6.107409</td>\n",
       "      <td>4.091625</td>\n",
       "      <td>3.248409</td>\n",
       "      <td>...</td>\n",
       "      <td>15.402841</td>\n",
       "      <td>18.166068</td>\n",
       "      <td>16.566273</td>\n",
       "      <td>16.050568</td>\n",
       "      <td>7.585966</td>\n",
       "      <td>0.185341</td>\n",
       "      <td>5.608455</td>\n",
       "      <td>5.310705</td>\n",
       "      <td>8.514114</td>\n",
       "      <td>5.644477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4537</th>\n",
       "      <td>3.140114</td>\n",
       "      <td>2.355761</td>\n",
       "      <td>2.947682</td>\n",
       "      <td>2.832591</td>\n",
       "      <td>3.312102</td>\n",
       "      <td>3.492955</td>\n",
       "      <td>0.493170</td>\n",
       "      <td>2.832591</td>\n",
       "      <td>2.075545</td>\n",
       "      <td>1.687136</td>\n",
       "      <td>...</td>\n",
       "      <td>2.699602</td>\n",
       "      <td>2.480773</td>\n",
       "      <td>3.833159</td>\n",
       "      <td>4.187477</td>\n",
       "      <td>2.078386</td>\n",
       "      <td>5.251841</td>\n",
       "      <td>2.068795</td>\n",
       "      <td>2.252273</td>\n",
       "      <td>2.667909</td>\n",
       "      <td>2.605636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4538</th>\n",
       "      <td>3.907591</td>\n",
       "      <td>3.279375</td>\n",
       "      <td>1.924352</td>\n",
       "      <td>1.674227</td>\n",
       "      <td>2.337955</td>\n",
       "      <td>2.564091</td>\n",
       "      <td>0.493170</td>\n",
       "      <td>3.134273</td>\n",
       "      <td>2.337159</td>\n",
       "      <td>2.073455</td>\n",
       "      <td>...</td>\n",
       "      <td>2.365898</td>\n",
       "      <td>2.209568</td>\n",
       "      <td>3.826091</td>\n",
       "      <td>4.673398</td>\n",
       "      <td>0.526318</td>\n",
       "      <td>2.274341</td>\n",
       "      <td>2.937318</td>\n",
       "      <td>2.974318</td>\n",
       "      <td>1.473125</td>\n",
       "      <td>2.510682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4541</th>\n",
       "      <td>6.722045</td>\n",
       "      <td>7.792432</td>\n",
       "      <td>7.052114</td>\n",
       "      <td>7.634705</td>\n",
       "      <td>11.214114</td>\n",
       "      <td>12.201534</td>\n",
       "      <td>9.317886</td>\n",
       "      <td>15.939886</td>\n",
       "      <td>8.500318</td>\n",
       "      <td>6.301864</td>\n",
       "      <td>...</td>\n",
       "      <td>3.204239</td>\n",
       "      <td>3.136705</td>\n",
       "      <td>5.142432</td>\n",
       "      <td>5.929659</td>\n",
       "      <td>4.157114</td>\n",
       "      <td>5.485727</td>\n",
       "      <td>4.972614</td>\n",
       "      <td>6.763159</td>\n",
       "      <td>3.852273</td>\n",
       "      <td>4.061705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4542</th>\n",
       "      <td>1.853955</td>\n",
       "      <td>2.153068</td>\n",
       "      <td>3.405818</td>\n",
       "      <td>3.305409</td>\n",
       "      <td>0.988386</td>\n",
       "      <td>1.269386</td>\n",
       "      <td>1.456500</td>\n",
       "      <td>2.549125</td>\n",
       "      <td>1.707114</td>\n",
       "      <td>2.132795</td>\n",
       "      <td>...</td>\n",
       "      <td>1.403568</td>\n",
       "      <td>2.225841</td>\n",
       "      <td>3.416227</td>\n",
       "      <td>3.743045</td>\n",
       "      <td>1.738102</td>\n",
       "      <td>2.136250</td>\n",
       "      <td>0.566989</td>\n",
       "      <td>0.604614</td>\n",
       "      <td>1.684659</td>\n",
       "      <td>2.182705</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3055 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GSM1004655  GSM1004656  GSM1004657  GSM1004658  GSM1004659  GSM1004660  \\\n",
       "ID                                                                             \n",
       "3       1.437273    1.674227    5.305341    4.193955    1.227295    3.416227   \n",
       "4       2.049591    2.043773    4.909659    4.769750    1.303795    2.549125   \n",
       "6       6.138932    5.641727   11.108727   10.532295    4.600114    9.868205   \n",
       "7       1.655091    1.115705    6.016773    6.079091    1.459250    2.676705   \n",
       "9       2.207614    2.803170    4.166545    3.309591    0.425273    0.958455   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "4536    6.042318    5.310705    8.327068    8.044136    5.652591    5.437250   \n",
       "4537    3.140114    2.355761    2.947682    2.832591    3.312102    3.492955   \n",
       "4538    3.907591    3.279375    1.924352    1.674227    2.337955    2.564091   \n",
       "4541    6.722045    7.792432    7.052114    7.634705   11.214114   12.201534   \n",
       "4542    1.853955    2.153068    3.405818    3.305409    0.988386    1.269386   \n",
       "\n",
       "      GSM1004661  GSM1004662  GSM1004663  GSM1004664  ...  GSM1004689  \\\n",
       "ID                                                    ...               \n",
       "3       3.718273    4.329773    2.180568    1.933909  ...    2.263227   \n",
       "4       5.019545    5.817341    2.505500    2.462614  ...    2.949727   \n",
       "6      10.034773   12.374955    6.749909    6.394932  ...    9.937273   \n",
       "7       1.828693    2.787205    2.109114    2.007932  ...    2.164375   \n",
       "9       1.397477    2.026455    1.268511    1.314568  ...    1.655739   \n",
       "...          ...         ...         ...         ...  ...         ...   \n",
       "4536    3.649705    6.107409    4.091625    3.248409  ...   15.402841   \n",
       "4537    0.493170    2.832591    2.075545    1.687136  ...    2.699602   \n",
       "4538    0.493170    3.134273    2.337159    2.073455  ...    2.365898   \n",
       "4541    9.317886   15.939886    8.500318    6.301864  ...    3.204239   \n",
       "4542    1.456500    2.549125    1.707114    2.132795  ...    1.403568   \n",
       "\n",
       "      GSM1004690  GSM1004691  GSM1004692  GSM1004693  GSM1004694  GSM1004695  \\\n",
       "ID                                                                             \n",
       "3       1.810614    5.394136    5.185057    2.817341    2.631955    1.718864   \n",
       "4       2.252273    8.628886    9.682500    3.649705    3.624045    0.911318   \n",
       "6       9.117500   12.210477   14.781477   10.952080   10.515000   10.150250   \n",
       "7       1.674795    4.452432    5.394136    2.623705    2.234193    1.003591   \n",
       "9       1.375727    3.354136    4.187477    2.204364    1.943034    1.316227   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "4536   18.166068   16.566273   16.050568    7.585966    0.185341    5.608455   \n",
       "4537    2.480773    3.833159    4.187477    2.078386    5.251841    2.068795   \n",
       "4538    2.209568    3.826091    4.673398    0.526318    2.274341    2.937318   \n",
       "4541    3.136705    5.142432    5.929659    4.157114    5.485727    4.972614   \n",
       "4542    2.225841    3.416227    3.743045    1.738102    2.136250    0.566989   \n",
       "\n",
       "      GSM1004696  GSM1004697  GSM1004698  \n",
       "ID                                        \n",
       "3       0.911318    1.290841    1.184500  \n",
       "4       1.885591    3.894159    2.964523  \n",
       "6       8.656409    5.062489    4.377955  \n",
       "7       1.744398    1.441341    1.398080  \n",
       "9       1.314568    0.830182    1.011409  \n",
       "...          ...         ...         ...  \n",
       "4536    5.310705    8.514114    5.644477  \n",
       "4537    2.252273    2.667909    2.605636  \n",
       "4538    2.974318    1.473125    2.510682  \n",
       "4541    6.763159    3.852273    4.061705  \n",
       "4542    0.604614    1.684659    2.182705  \n",
       "\n",
       "[3055 rows x 44 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['910', '4260', '1981', '2381', '4288'], 'GB_ACC': ['BE833259', 'BE702227', 'BF364095', 'BE081005', 'AW880607'], 'SPOT_TYPE': ['Exonic', 'Exonic', 'Exonic', 'Exonic', 'Exonic'], 'GENE_ID': [85439.0, 2776.0, 84131.0, 2776.0, 54768.0], 'GENE_SYMBOL': ['STON2', 'GNAQ', 'CEP78', 'GNAQ', 'HYDIN'], 'GENE_ANNOTATION': ['stonin 2', 'Guanine nucleotide binding protein (G protein), q polypeptide', 'centrosomal protein 78kDa', 'Guanine nucleotide binding protein (G protein), q polypeptide', 'hydrocephalus inducing homolog 2 (mouse); hydrocephalus inducing homolog (mouse)'], 'CPC_CODING_POTENTIAL': ['noncoding', 'noncoding', 'noncoding', 'noncoding', '-'], 'SEQUENCE': ['CTGATCCGCTTAAGCTTAGTATGTTTGAGTGTGTAATTTTAGTTTCTTTTCTGGTTGTATTTGTGGTAGTCAGATGTGTTGGATTGATTCCAACTGGACAGAGTAAGGAATTCCAGCATCCTCTTCCTGCTTGCTCGTGTTACCCCACAGATCAAACCCTCAATTCTAGTTGGGGATGCTGTCTAGCCCCACACCATGACTGAAGCCTTAAGCACTGTTGCGCCTCCATGTGCTTTGGATCAGCAACCCCAGTGGTATTCTACCAGAGCATTGTGGGAAAGCAGATGTATAGTCAGGTCCCAACAGCAAATTGTTGGGTGTGAGAGTTCTAAAGTATAGGGGTGAGGGAAGAGAAGGATATGAACTCCT', 'CTCTTCCGAAAGATATATCTTGGTTAGAAACACAAAAAAATAAAACTAGTAATATTGTATGTTTATCTATCTCTACATATTTCCAGCATATGTAGCGTTAATAGATCTGTCCTGGTAACTGTGTCTTTGGGATTTCATTTTGGTTCCATCAAATTAGGAAAAGAAATGGCTTAGTTGTATATGATTAGCTAGAGATTTTTGGAGCCAGACACCTGCTGTTTAGTAGATAACTTAGTACAGACCCTAAACTTGTCATTTGTTTTTCTCACAGAATAGCCATTTCCTGCTGTCTTCCCAATGATCACTGCCCTTTCAATAACACTCTTGCCTCTAGAATCATATG', 'CCTTTGAAATGACTGGAGAATATTAAAATAAGAAATAATCATGCAGAGTTGGAAACCAGAAATCTGAACAGTGAAATTGTCTGGCAGGATAAGACGCAGATGCATTTAAGTACCAGTTCAATTAAAGGATGGAACAGCTAAGCCATTCCACTCATCTTCGTGAGCATCTGATTCTGGAGTTTGCGCACCGAGGCTAAGAAAGCAGCTATCTGAAGTGGGAGCGCTGACCCAAGAAATGCTGGGATCGGAGAATAAGGGAATTATCCAAAATGGCTCCGAAGAGGAACTGAAGTTAAGCTGCCCACATGATCTCTCTAACTATGATGACCTGCCACTTCCGTTTATAATCACCACATAAGTGCCTGTAATCATTTGTGTTCATTAAAAGTGAACCAGAATTCCCATTTGGATGAAAAAATAACACTTCCAACTTTAATCTTAGGCCCTCATTTATAAATATGGACAACCAAGAATCATCAAATTTGAAGAAAACCAGTAACATAAAAGGAGGCATGAAATTAAAATTAACCTGTTCAAGAAGATAGTTACTAGGAGAAACATGAAATTTTTAAATTAATGAATCAAAATCTTCAGCAATTCATAAAGATACTGTGTTCATAAAGAATAGGATGCCATGACAAAAATATTTCGAGTTTCCTGGAATTAAACATTTGA', 'CCGTAGCACTTCCTGTACTATACAAGAACAAGAACATAAAACACAGAAACCTTTCTTCAGCATACCAAGGCAAGCAGCCATTTCATGACTCACTTAACACATTGCAGTGTACCAGTTTACAGATGATTTTTCCCTTTTTGCGTGACATGGCAGACCCTGCCGCCAGAGAATTCCTTATTTGTAAATTGGAAGTTTCTACTATGCCTTACAGAGCTTAAATTCAGAAGTTTGTGCCTCATATCTGAAACAAAGGGAAATAACACACCCATTCAAAAGTAAATAAATCTCCTATAAGTTTTTGTTTTTAACATTTCCATATAAAGAGCTCTGTTGAATGTCATGAATAGACTGGAACATAACATTTTAAGAACCTGCATATGTTGTTTACTAGCAGATGACAACTACAAAAGGAATCTGAAGAACACGTAAAACTTGTATTTTTTGGCTTGCACGGGGATATCAACTACCAGGCCTTTTTCAAAAAGGTATTTCAGCTAAGGGGCCAATACACTTTTTGGTACTCCAGACTATCTTGATTGTCAAGGTGTCCGACCTGTATTTTTAAATTTTATACTGCCACATGATTGTA', 'GAGGGATTGGCCCCTGTGGGTCAAATCTCACTTCAAATATTTCCGTTTCACAATGAGGCAGATTCTTTACACGATCTAGCTCAGTACTGAATCCTGTCTCATGAAGGACACGCTTGTCTGCATGGAATGACACTGGAAAGTGACTGGTGTTGATGATCTTGATGATGTGGGTTCGGACTTCGCCAAGGATGATGTAGCCAAAGTCCAGGATGTACTCTGGTAGCTGGATTTTGGCCAGTTTGCGGCGACTCCGATGGCTGAAGCAGGGGTCATCCATAGGATCAGGGGTGGTTGTATTCTGATGTTCTAGGACATAGCTTTGGACTATAAGTCTTTCTACCTCCATCTGGAGATGAGCACTTACCTCAGCAGGCTCGTCTTCTGGCACTTCCTCAGTTATTACGTCAAAGTGATCGAGCATTTCACATTTGTTATACTCTTTGTCTGTGTTTTTCCTGGCTTGATTCAAGAACATTTCATACTTTTCATTTGCTGTGAGGTTCCTGGGGAGATCGAGGCAGATTTGG'], 'COORDINATES': ['chr14:81727450-81727801', 'chr9:80331795-80332137', 'chr9:80885760-80886427', 'chr9:80332512-80333071', 'chr1_gl000192_random:211847-219091'], 'CLONE ID': ['QV3-OT0065-150600-231-c01', 'QV0-NN1022-070700-294-f10', 'QV0-NN1022-120500-220-f07', 'QV1-BT0631-210300-120-f05', 'QV0-OT0031-100300-157-h12'], 'SPOT_ID': ['Exonic', 'Exonic', 'Exonic', 'Exonic', 'Exonic']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['3', '4', '6', '7', '9', '10', '15', '16', '17', '19', '20', '21', '22', '23', '26', '27', '31', '32', '33', '35']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['910', '4260', '1981', '2381', '4288'], 'GB_ACC': ['BE833259', 'BE702227', 'BF364095', 'BE081005', 'AW880607'], 'SPOT_TYPE': ['Exonic', 'Exonic', 'Exonic', 'Exonic', 'Exonic'], 'GENE_ID': [85439.0, 2776.0, 84131.0, 2776.0, 54768.0], 'GENE_SYMBOL': ['STON2', 'GNAQ', 'CEP78', 'GNAQ', 'HYDIN'], 'GENE_ANNOTATION': ['stonin 2', 'Guanine nucleotide binding protein (G protein), q polypeptide', 'centrosomal protein 78kDa', 'Guanine nucleotide binding protein (G protein), q polypeptide', 'hydrocephalus inducing homolog 2 (mouse); hydrocephalus inducing homolog (mouse)'], 'CPC_CODING_POTENTIAL': ['noncoding', 'noncoding', 'noncoding', 'noncoding', '-'], 'SEQUENCE': ['CTGATCCGCTTAAGCTTAGTATGTTTGAGTGTGTAATTTTAGTTTCTTTTCTGGTTGTATTTGTGGTAGTCAGATGTGTTGGATTGATTCCAACTGGACAGAGTAAGGAATTCCAGCATCCTCTTCCTGCTTGCTCGTGTTACCCCACAGATCAAACCCTCAATTCTAGTTGGGGATGCTGTCTAGCCCCACACCATGACTGAAGCCTTAAGCACTGTTGCGCCTCCATGTGCTTTGGATCAGCAACCCCAGTGGTATTCTACCAGAGCATTGTGGGAAAGCAGATGTATAGTCAGGTCCCAACAGCAAATTGTTGGGTGTGAGAGTTCTAAAGTATAGGGGTGAGGGAAGAGAAGGATATGAACTCCT', 'CTCTTCCGAAAGATATATCTTGGTTAGAAACACAAAAAAATAAAACTAGTAATATTGTATGTTTATCTATCTCTACATATTTCCAGCATATGTAGCGTTAATAGATCTGTCCTGGTAACTGTGTCTTTGGGATTTCATTTTGGTTCCATCAAATTAGGAAAAGAAATGGCTTAGTTGTATATGATTAGCTAGAGATTTTTGGAGCCAGACACCTGCTGTTTAGTAGATAACTTAGTACAGACCCTAAACTTGTCATTTGTTTTTCTCACAGAATAGCCATTTCCTGCTGTCTTCCCAATGATCACTGCCCTTTCAATAACACTCTTGCCTCTAGAATCATATG', 'CCTTTGAAATGACTGGAGAATATTAAAATAAGAAATAATCATGCAGAGTTGGAAACCAGAAATCTGAACAGTGAAATTGTCTGGCAGGATAAGACGCAGATGCATTTAAGTACCAGTTCAATTAAAGGATGGAACAGCTAAGCCATTCCACTCATCTTCGTGAGCATCTGATTCTGGAGTTTGCGCACCGAGGCTAAGAAAGCAGCTATCTGAAGTGGGAGCGCTGACCCAAGAAATGCTGGGATCGGAGAATAAGGGAATTATCCAAAATGGCTCCGAAGAGGAACTGAAGTTAAGCTGCCCACATGATCTCTCTAACTATGATGACCTGCCACTTCCGTTTATAATCACCACATAAGTGCCTGTAATCATTTGTGTTCATTAAAAGTGAACCAGAATTCCCATTTGGATGAAAAAATAACACTTCCAACTTTAATCTTAGGCCCTCATTTATAAATATGGACAACCAAGAATCATCAAATTTGAAGAAAACCAGTAACATAAAAGGAGGCATGAAATTAAAATTAACCTGTTCAAGAAGATAGTTACTAGGAGAAACATGAAATTTTTAAATTAATGAATCAAAATCTTCAGCAATTCATAAAGATACTGTGTTCATAAAGAATAGGATGCCATGACAAAAATATTTCGAGTTTCCTGGAATTAAACATTTGA', 'CCGTAGCACTTCCTGTACTATACAAGAACAAGAACATAAAACACAGAAACCTTTCTTCAGCATACCAAGGCAAGCAGCCATTTCATGACTCACTTAACACATTGCAGTGTACCAGTTTACAGATGATTTTTCCCTTTTTGCGTGACATGGCAGACCCTGCCGCCAGAGAATTCCTTATTTGTAAATTGGAAGTTTCTACTATGCCTTACAGAGCTTAAATTCAGAAGTTTGTGCCTCATATCTGAAACAAAGGGAAATAACACACCCATTCAAAAGTAAATAAATCTCCTATAAGTTTTTGTTTTTAACATTTCCATATAAAGAGCTCTGTTGAATGTCATGAATAGACTGGAACATAACATTTTAAGAACCTGCATATGTTGTTTACTAGCAGATGACAACTACAAAAGGAATCTGAAGAACACGTAAAACTTGTATTTTTTGGCTTGCACGGGGATATCAACTACCAGGCCTTTTTCAAAAAGGTATTTCAGCTAAGGGGCCAATACACTTTTTGGTACTCCAGACTATCTTGATTGTCAAGGTGTCCGACCTGTATTTTTAAATTTTATACTGCCACATGATTGTA', 'GAGGGATTGGCCCCTGTGGGTCAAATCTCACTTCAAATATTTCCGTTTCACAATGAGGCAGATTCTTTACACGATCTAGCTCAGTACTGAATCCTGTCTCATGAAGGACACGCTTGTCTGCATGGAATGACACTGGAAAGTGACTGGTGTTGATGATCTTGATGATGTGGGTTCGGACTTCGCCAAGGATGATGTAGCCAAAGTCCAGGATGTACTCTGGTAGCTGGATTTTGGCCAGTTTGCGGCGACTCCGATGGCTGAAGCAGGGGTCATCCATAGGATCAGGGGTGGTTGTATTCTGATGTTCTAGGACATAGCTTTGGACTATAAGTCTTTCTACCTCCATCTGGAGATGAGCACTTACCTCAGCAGGCTCGTCTTCTGGCACTTCCTCAGTTATTACGTCAAAGTGATCGAGCATTTCACATTTGTTATACTCTTTGTCTGTGTTTTTCCTGGCTTGATTCAAGAACATTTCATACTTTTCATTTGCTGTGAGGTTCCTGGGGAGATCGAGGCAGATTTGG'], 'COORDINATES': ['chr14:81727450-81727801', 'chr9:80331795-80332137', 'chr9:80885760-80886427', 'chr9:80332512-80333071', 'chr1_gl000192_random:211847-219091'], 'CLONE ID': ['QV3-OT0065-150600-231-c01', 'QV0-NN1022-070700-294-f10', 'QV0-NN1022-120500-220-f07', 'QV1-BT0631-210300-120-f05', 'QV0-OT0031-100300-157-h12'], 'SPOT_ID': ['Exonic', 'Exonic', 'Exonic', 'Exonic', 'Exonic']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_TYPE', 'GENE_ID', 'GENE_SYMBOL',\n",
       "       'GENE_ANNOTATION', 'CPC_CODING_POTENTIAL', 'SEQUENCE', 'COORDINATES',\n",
       "       'CLONE ID', 'SPOT_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GENE_SYMBOL'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1004655</th>\n",
       "      <th>GSM1004656</th>\n",
       "      <th>GSM1004657</th>\n",
       "      <th>GSM1004658</th>\n",
       "      <th>GSM1004659</th>\n",
       "      <th>GSM1004660</th>\n",
       "      <th>GSM1004661</th>\n",
       "      <th>GSM1004662</th>\n",
       "      <th>GSM1004663</th>\n",
       "      <th>GSM1004664</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1004689</th>\n",
       "      <th>GSM1004690</th>\n",
       "      <th>GSM1004691</th>\n",
       "      <th>GSM1004692</th>\n",
       "      <th>GSM1004693</th>\n",
       "      <th>GSM1004694</th>\n",
       "      <th>GSM1004695</th>\n",
       "      <th>GSM1004696</th>\n",
       "      <th>GSM1004697</th>\n",
       "      <th>GSM1004698</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>11.611000</td>\n",
       "      <td>12.398614</td>\n",
       "      <td>8.044136</td>\n",
       "      <td>8.556591</td>\n",
       "      <td>7.704807</td>\n",
       "      <td>6.522909</td>\n",
       "      <td>9.999795</td>\n",
       "      <td>9.772500</td>\n",
       "      <td>7.132091</td>\n",
       "      <td>5.202318</td>\n",
       "      <td>...</td>\n",
       "      <td>8.696045</td>\n",
       "      <td>7.371818</td>\n",
       "      <td>8.591148</td>\n",
       "      <td>8.100955</td>\n",
       "      <td>11.280591</td>\n",
       "      <td>11.148239</td>\n",
       "      <td>11.108727</td>\n",
       "      <td>9.756568</td>\n",
       "      <td>8.489977</td>\n",
       "      <td>5.909795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARSD1</th>\n",
       "      <td>5.463477</td>\n",
       "      <td>7.561534</td>\n",
       "      <td>4.762284</td>\n",
       "      <td>4.645347</td>\n",
       "      <td>8.187159</td>\n",
       "      <td>4.843648</td>\n",
       "      <td>5.484097</td>\n",
       "      <td>6.125852</td>\n",
       "      <td>4.627341</td>\n",
       "      <td>4.554318</td>\n",
       "      <td>...</td>\n",
       "      <td>5.583023</td>\n",
       "      <td>4.967341</td>\n",
       "      <td>5.021307</td>\n",
       "      <td>5.170159</td>\n",
       "      <td>4.953369</td>\n",
       "      <td>5.731830</td>\n",
       "      <td>5.867227</td>\n",
       "      <td>5.509068</td>\n",
       "      <td>6.256591</td>\n",
       "      <td>4.950045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATF</th>\n",
       "      <td>5.866136</td>\n",
       "      <td>6.779500</td>\n",
       "      <td>7.152886</td>\n",
       "      <td>7.174977</td>\n",
       "      <td>5.923523</td>\n",
       "      <td>7.814091</td>\n",
       "      <td>5.647227</td>\n",
       "      <td>5.936205</td>\n",
       "      <td>5.132614</td>\n",
       "      <td>8.291773</td>\n",
       "      <td>...</td>\n",
       "      <td>6.651886</td>\n",
       "      <td>6.834841</td>\n",
       "      <td>5.958523</td>\n",
       "      <td>6.061705</td>\n",
       "      <td>6.756705</td>\n",
       "      <td>6.643341</td>\n",
       "      <td>8.088909</td>\n",
       "      <td>6.793795</td>\n",
       "      <td>6.476136</td>\n",
       "      <td>7.770227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA3</th>\n",
       "      <td>2.304568</td>\n",
       "      <td>2.531636</td>\n",
       "      <td>4.169409</td>\n",
       "      <td>3.884727</td>\n",
       "      <td>0.992409</td>\n",
       "      <td>2.001591</td>\n",
       "      <td>2.025750</td>\n",
       "      <td>4.709955</td>\n",
       "      <td>2.187364</td>\n",
       "      <td>2.650977</td>\n",
       "      <td>...</td>\n",
       "      <td>4.423670</td>\n",
       "      <td>4.664216</td>\n",
       "      <td>5.718386</td>\n",
       "      <td>6.238136</td>\n",
       "      <td>4.003159</td>\n",
       "      <td>5.027864</td>\n",
       "      <td>2.009568</td>\n",
       "      <td>0.258432</td>\n",
       "      <td>2.962682</td>\n",
       "      <td>2.970159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB10</th>\n",
       "      <td>3.128023</td>\n",
       "      <td>4.258580</td>\n",
       "      <td>5.121580</td>\n",
       "      <td>5.776841</td>\n",
       "      <td>5.195102</td>\n",
       "      <td>6.675045</td>\n",
       "      <td>5.629682</td>\n",
       "      <td>5.837705</td>\n",
       "      <td>6.154091</td>\n",
       "      <td>6.749909</td>\n",
       "      <td>...</td>\n",
       "      <td>1.439886</td>\n",
       "      <td>2.184136</td>\n",
       "      <td>1.479148</td>\n",
       "      <td>1.606227</td>\n",
       "      <td>4.489114</td>\n",
       "      <td>4.833023</td>\n",
       "      <td>5.506409</td>\n",
       "      <td>3.971659</td>\n",
       "      <td>3.641545</td>\n",
       "      <td>3.018000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>2.842023</td>\n",
       "      <td>5.670727</td>\n",
       "      <td>3.114318</td>\n",
       "      <td>3.013386</td>\n",
       "      <td>7.037091</td>\n",
       "      <td>2.927023</td>\n",
       "      <td>4.182432</td>\n",
       "      <td>4.116295</td>\n",
       "      <td>3.781341</td>\n",
       "      <td>4.459932</td>\n",
       "      <td>...</td>\n",
       "      <td>3.460523</td>\n",
       "      <td>3.260307</td>\n",
       "      <td>2.971625</td>\n",
       "      <td>3.644750</td>\n",
       "      <td>2.470284</td>\n",
       "      <td>3.244091</td>\n",
       "      <td>4.407591</td>\n",
       "      <td>3.291898</td>\n",
       "      <td>3.743045</td>\n",
       "      <td>2.431773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>1.811693</td>\n",
       "      <td>2.781568</td>\n",
       "      <td>2.068795</td>\n",
       "      <td>2.061159</td>\n",
       "      <td>2.292614</td>\n",
       "      <td>2.279068</td>\n",
       "      <td>2.309682</td>\n",
       "      <td>3.003500</td>\n",
       "      <td>3.627909</td>\n",
       "      <td>2.482614</td>\n",
       "      <td>...</td>\n",
       "      <td>1.370273</td>\n",
       "      <td>2.416659</td>\n",
       "      <td>1.479148</td>\n",
       "      <td>2.442682</td>\n",
       "      <td>2.807750</td>\n",
       "      <td>4.364205</td>\n",
       "      <td>3.128023</td>\n",
       "      <td>3.073705</td>\n",
       "      <td>2.769273</td>\n",
       "      <td>2.724955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARL5A</th>\n",
       "      <td>5.899591</td>\n",
       "      <td>5.942318</td>\n",
       "      <td>5.942318</td>\n",
       "      <td>6.585159</td>\n",
       "      <td>6.675045</td>\n",
       "      <td>6.543614</td>\n",
       "      <td>1.284795</td>\n",
       "      <td>2.230034</td>\n",
       "      <td>3.418818</td>\n",
       "      <td>2.890795</td>\n",
       "      <td>...</td>\n",
       "      <td>3.263795</td>\n",
       "      <td>3.665227</td>\n",
       "      <td>4.000250</td>\n",
       "      <td>4.701227</td>\n",
       "      <td>7.104273</td>\n",
       "      <td>7.253159</td>\n",
       "      <td>8.872409</td>\n",
       "      <td>9.100227</td>\n",
       "      <td>4.407591</td>\n",
       "      <td>3.595500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NONO</th>\n",
       "      <td>18.844500</td>\n",
       "      <td>18.554023</td>\n",
       "      <td>21.731909</td>\n",
       "      <td>21.803500</td>\n",
       "      <td>14.853091</td>\n",
       "      <td>15.429045</td>\n",
       "      <td>18.751341</td>\n",
       "      <td>17.925205</td>\n",
       "      <td>21.613500</td>\n",
       "      <td>21.613500</td>\n",
       "      <td>...</td>\n",
       "      <td>15.005955</td>\n",
       "      <td>14.472636</td>\n",
       "      <td>22.955568</td>\n",
       "      <td>21.140386</td>\n",
       "      <td>17.286795</td>\n",
       "      <td>16.766295</td>\n",
       "      <td>19.003841</td>\n",
       "      <td>20.658341</td>\n",
       "      <td>17.032136</td>\n",
       "      <td>19.342568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKRIP1</th>\n",
       "      <td>3.841659</td>\n",
       "      <td>8.121682</td>\n",
       "      <td>3.080023</td>\n",
       "      <td>3.413045</td>\n",
       "      <td>8.044136</td>\n",
       "      <td>5.078136</td>\n",
       "      <td>3.000057</td>\n",
       "      <td>4.526136</td>\n",
       "      <td>3.093068</td>\n",
       "      <td>3.139205</td>\n",
       "      <td>...</td>\n",
       "      <td>4.706273</td>\n",
       "      <td>7.139318</td>\n",
       "      <td>3.619250</td>\n",
       "      <td>4.172500</td>\n",
       "      <td>2.942284</td>\n",
       "      <td>5.166523</td>\n",
       "      <td>7.104273</td>\n",
       "      <td>6.187409</td>\n",
       "      <td>5.557045</td>\n",
       "      <td>9.539614</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2273 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1004655  GSM1004656  GSM1004657  GSM1004658  GSM1004659  \\\n",
       "A2ML1     11.611000   12.398614    8.044136    8.556591    7.704807   \n",
       "AARSD1     5.463477    7.561534    4.762284    4.645347    8.187159   \n",
       "AATF       5.866136    6.779500    7.152886    7.174977    5.923523   \n",
       "ABCA3      2.304568    2.531636    4.169409    3.884727    0.992409   \n",
       "ABCB10     3.128023    4.258580    5.121580    5.776841    5.195102   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZXDC       2.842023    5.670727    3.114318    3.013386    7.037091   \n",
       "ZZEF1      1.811693    2.781568    2.068795    2.061159    2.292614   \n",
       "ARL5A      5.899591    5.942318    5.942318    6.585159    6.675045   \n",
       "NONO      18.844500   18.554023   21.731909   21.803500   14.853091   \n",
       "PRKRIP1    3.841659    8.121682    3.080023    3.413045    8.044136   \n",
       "\n",
       "         GSM1004660  GSM1004661  GSM1004662  GSM1004663  GSM1004664  ...  \\\n",
       "A2ML1      6.522909    9.999795    9.772500    7.132091    5.202318  ...   \n",
       "AARSD1     4.843648    5.484097    6.125852    4.627341    4.554318  ...   \n",
       "AATF       7.814091    5.647227    5.936205    5.132614    8.291773  ...   \n",
       "ABCA3      2.001591    2.025750    4.709955    2.187364    2.650977  ...   \n",
       "ABCB10     6.675045    5.629682    5.837705    6.154091    6.749909  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "ZXDC       2.927023    4.182432    4.116295    3.781341    4.459932  ...   \n",
       "ZZEF1      2.279068    2.309682    3.003500    3.627909    2.482614  ...   \n",
       "ARL5A      6.543614    1.284795    2.230034    3.418818    2.890795  ...   \n",
       "NONO      15.429045   18.751341   17.925205   21.613500   21.613500  ...   \n",
       "PRKRIP1    5.078136    3.000057    4.526136    3.093068    3.139205  ...   \n",
       "\n",
       "         GSM1004689  GSM1004690  GSM1004691  GSM1004692  GSM1004693  \\\n",
       "A2ML1      8.696045    7.371818    8.591148    8.100955   11.280591   \n",
       "AARSD1     5.583023    4.967341    5.021307    5.170159    4.953369   \n",
       "AATF       6.651886    6.834841    5.958523    6.061705    6.756705   \n",
       "ABCA3      4.423670    4.664216    5.718386    6.238136    4.003159   \n",
       "ABCB10     1.439886    2.184136    1.479148    1.606227    4.489114   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "ZXDC       3.460523    3.260307    2.971625    3.644750    2.470284   \n",
       "ZZEF1      1.370273    2.416659    1.479148    2.442682    2.807750   \n",
       "ARL5A      3.263795    3.665227    4.000250    4.701227    7.104273   \n",
       "NONO      15.005955   14.472636   22.955568   21.140386   17.286795   \n",
       "PRKRIP1    4.706273    7.139318    3.619250    4.172500    2.942284   \n",
       "\n",
       "         GSM1004694  GSM1004695  GSM1004696  GSM1004697  GSM1004698  \n",
       "A2ML1     11.148239   11.108727    9.756568    8.489977    5.909795  \n",
       "AARSD1     5.731830    5.867227    5.509068    6.256591    4.950045  \n",
       "AATF       6.643341    8.088909    6.793795    6.476136    7.770227  \n",
       "ABCA3      5.027864    2.009568    0.258432    2.962682    2.970159  \n",
       "ABCB10     4.833023    5.506409    3.971659    3.641545    3.018000  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "ZXDC       3.244091    4.407591    3.291898    3.743045    2.431773  \n",
       "ZZEF1      4.364205    3.128023    3.073705    2.769273    2.724955  \n",
       "ARL5A      7.253159    8.872409    9.100227    4.407591    3.595500  \n",
       "NONO      16.766295   19.003841   20.658341   17.032136   19.342568  \n",
       "PRKRIP1    5.166523    7.104273    6.187409    5.557045    9.539614  \n",
       "\n",
       "[2273 rows x 44 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>AARSD1</th>\n",
       "      <th>AATF</th>\n",
       "      <th>ABCA3</th>\n",
       "      <th>ABCB10</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCB7</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF841</th>\n",
       "      <th>ZNHIT3</th>\n",
       "      <th>ZRANB3</th>\n",
       "      <th>ZSCAN18</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ARL5A</th>\n",
       "      <th>NONO</th>\n",
       "      <th>PRKRIP1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1004655</th>\n",
       "      <td>1.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.611000</td>\n",
       "      <td>5.463477</td>\n",
       "      <td>5.866136</td>\n",
       "      <td>2.304568</td>\n",
       "      <td>3.128023</td>\n",
       "      <td>2.288523</td>\n",
       "      <td>3.816227</td>\n",
       "      <td>...</td>\n",
       "      <td>52.649727</td>\n",
       "      <td>428.064205</td>\n",
       "      <td>5.563250</td>\n",
       "      <td>8.566886</td>\n",
       "      <td>2.134705</td>\n",
       "      <td>2.842023</td>\n",
       "      <td>1.811693</td>\n",
       "      <td>5.899591</td>\n",
       "      <td>18.844500</td>\n",
       "      <td>3.841659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004656</th>\n",
       "      <td>1.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.398614</td>\n",
       "      <td>7.561534</td>\n",
       "      <td>6.779500</td>\n",
       "      <td>2.531636</td>\n",
       "      <td>4.258580</td>\n",
       "      <td>1.857784</td>\n",
       "      <td>5.336545</td>\n",
       "      <td>...</td>\n",
       "      <td>53.024932</td>\n",
       "      <td>428.064205</td>\n",
       "      <td>4.555818</td>\n",
       "      <td>10.590000</td>\n",
       "      <td>1.646068</td>\n",
       "      <td>5.670727</td>\n",
       "      <td>2.781568</td>\n",
       "      <td>5.942318</td>\n",
       "      <td>18.554023</td>\n",
       "      <td>8.121682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004657</th>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.044136</td>\n",
       "      <td>4.762284</td>\n",
       "      <td>7.152886</td>\n",
       "      <td>4.169409</td>\n",
       "      <td>5.121580</td>\n",
       "      <td>2.802159</td>\n",
       "      <td>5.929659</td>\n",
       "      <td>...</td>\n",
       "      <td>56.879750</td>\n",
       "      <td>492.838000</td>\n",
       "      <td>6.010773</td>\n",
       "      <td>7.792432</td>\n",
       "      <td>3.739795</td>\n",
       "      <td>3.114318</td>\n",
       "      <td>2.068795</td>\n",
       "      <td>5.942318</td>\n",
       "      <td>21.731909</td>\n",
       "      <td>3.080023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004658</th>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.556591</td>\n",
       "      <td>4.645347</td>\n",
       "      <td>7.174977</td>\n",
       "      <td>3.884727</td>\n",
       "      <td>5.776841</td>\n",
       "      <td>3.129000</td>\n",
       "      <td>5.627193</td>\n",
       "      <td>...</td>\n",
       "      <td>53.339636</td>\n",
       "      <td>492.838000</td>\n",
       "      <td>6.126432</td>\n",
       "      <td>8.121682</td>\n",
       "      <td>3.108023</td>\n",
       "      <td>3.013386</td>\n",
       "      <td>2.061159</td>\n",
       "      <td>6.585159</td>\n",
       "      <td>21.803500</td>\n",
       "      <td>3.413045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004659</th>\n",
       "      <td>1.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.704807</td>\n",
       "      <td>8.187159</td>\n",
       "      <td>5.923523</td>\n",
       "      <td>0.992409</td>\n",
       "      <td>5.195102</td>\n",
       "      <td>1.369386</td>\n",
       "      <td>4.128000</td>\n",
       "      <td>...</td>\n",
       "      <td>61.150727</td>\n",
       "      <td>559.608205</td>\n",
       "      <td>4.766364</td>\n",
       "      <td>8.009045</td>\n",
       "      <td>1.378205</td>\n",
       "      <td>7.037091</td>\n",
       "      <td>2.292614</td>\n",
       "      <td>6.675045</td>\n",
       "      <td>14.853091</td>\n",
       "      <td>8.044136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004660</th>\n",
       "      <td>1.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.522909</td>\n",
       "      <td>4.843648</td>\n",
       "      <td>7.814091</td>\n",
       "      <td>2.001591</td>\n",
       "      <td>6.675045</td>\n",
       "      <td>3.592977</td>\n",
       "      <td>4.304909</td>\n",
       "      <td>...</td>\n",
       "      <td>57.500205</td>\n",
       "      <td>559.608205</td>\n",
       "      <td>4.132818</td>\n",
       "      <td>7.491523</td>\n",
       "      <td>2.551955</td>\n",
       "      <td>2.927023</td>\n",
       "      <td>2.279068</td>\n",
       "      <td>6.543614</td>\n",
       "      <td>15.429045</td>\n",
       "      <td>5.078136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004661</th>\n",
       "      <td>1.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.999795</td>\n",
       "      <td>5.484097</td>\n",
       "      <td>5.647227</td>\n",
       "      <td>2.025750</td>\n",
       "      <td>5.629682</td>\n",
       "      <td>4.240318</td>\n",
       "      <td>4.577591</td>\n",
       "      <td>...</td>\n",
       "      <td>62.049091</td>\n",
       "      <td>110.698523</td>\n",
       "      <td>4.736011</td>\n",
       "      <td>8.182250</td>\n",
       "      <td>2.531636</td>\n",
       "      <td>4.182432</td>\n",
       "      <td>2.309682</td>\n",
       "      <td>1.284795</td>\n",
       "      <td>18.751341</td>\n",
       "      <td>3.000057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004662</th>\n",
       "      <td>1.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.772500</td>\n",
       "      <td>6.125852</td>\n",
       "      <td>5.936205</td>\n",
       "      <td>4.709955</td>\n",
       "      <td>5.837705</td>\n",
       "      <td>3.957761</td>\n",
       "      <td>6.022886</td>\n",
       "      <td>...</td>\n",
       "      <td>52.649727</td>\n",
       "      <td>175.147705</td>\n",
       "      <td>4.940864</td>\n",
       "      <td>8.077614</td>\n",
       "      <td>2.713705</td>\n",
       "      <td>4.116295</td>\n",
       "      <td>3.003500</td>\n",
       "      <td>2.230034</td>\n",
       "      <td>17.925205</td>\n",
       "      <td>4.526136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004663</th>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.132091</td>\n",
       "      <td>4.627341</td>\n",
       "      <td>5.132614</td>\n",
       "      <td>2.187364</td>\n",
       "      <td>6.154091</td>\n",
       "      <td>1.673659</td>\n",
       "      <td>3.484932</td>\n",
       "      <td>...</td>\n",
       "      <td>61.539432</td>\n",
       "      <td>363.600432</td>\n",
       "      <td>6.819295</td>\n",
       "      <td>6.194659</td>\n",
       "      <td>1.844273</td>\n",
       "      <td>3.781341</td>\n",
       "      <td>3.627909</td>\n",
       "      <td>3.418818</td>\n",
       "      <td>21.613500</td>\n",
       "      <td>3.093068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004664</th>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.202318</td>\n",
       "      <td>4.554318</td>\n",
       "      <td>8.291773</td>\n",
       "      <td>2.650977</td>\n",
       "      <td>6.749909</td>\n",
       "      <td>3.503614</td>\n",
       "      <td>2.606432</td>\n",
       "      <td>...</td>\n",
       "      <td>53.883227</td>\n",
       "      <td>351.976114</td>\n",
       "      <td>7.990250</td>\n",
       "      <td>4.994000</td>\n",
       "      <td>1.340500</td>\n",
       "      <td>4.459932</td>\n",
       "      <td>2.482614</td>\n",
       "      <td>2.890795</td>\n",
       "      <td>21.613500</td>\n",
       "      <td>3.139205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004665</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.729409</td>\n",
       "      <td>5.508659</td>\n",
       "      <td>6.432477</td>\n",
       "      <td>2.658307</td>\n",
       "      <td>3.019966</td>\n",
       "      <td>0.893205</td>\n",
       "      <td>3.519932</td>\n",
       "      <td>...</td>\n",
       "      <td>65.250318</td>\n",
       "      <td>405.842364</td>\n",
       "      <td>9.504591</td>\n",
       "      <td>10.311523</td>\n",
       "      <td>3.673341</td>\n",
       "      <td>4.847295</td>\n",
       "      <td>2.849705</td>\n",
       "      <td>5.713614</td>\n",
       "      <td>22.483886</td>\n",
       "      <td>3.743045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004666</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.749909</td>\n",
       "      <td>4.967398</td>\n",
       "      <td>5.235159</td>\n",
       "      <td>4.222091</td>\n",
       "      <td>2.882489</td>\n",
       "      <td>1.199545</td>\n",
       "      <td>4.421409</td>\n",
       "      <td>...</td>\n",
       "      <td>68.054591</td>\n",
       "      <td>462.197295</td>\n",
       "      <td>4.605477</td>\n",
       "      <td>14.361295</td>\n",
       "      <td>3.524841</td>\n",
       "      <td>3.339045</td>\n",
       "      <td>3.824807</td>\n",
       "      <td>8.105477</td>\n",
       "      <td>16.566273</td>\n",
       "      <td>9.601432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004667</th>\n",
       "      <td>1.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.229159</td>\n",
       "      <td>4.116148</td>\n",
       "      <td>8.291773</td>\n",
       "      <td>2.534909</td>\n",
       "      <td>4.392636</td>\n",
       "      <td>2.126216</td>\n",
       "      <td>2.981818</td>\n",
       "      <td>...</td>\n",
       "      <td>52.142364</td>\n",
       "      <td>405.842364</td>\n",
       "      <td>7.095455</td>\n",
       "      <td>6.424318</td>\n",
       "      <td>2.035216</td>\n",
       "      <td>2.671386</td>\n",
       "      <td>3.575364</td>\n",
       "      <td>4.605477</td>\n",
       "      <td>14.588636</td>\n",
       "      <td>3.259227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004668</th>\n",
       "      <td>1.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.699705</td>\n",
       "      <td>5.179080</td>\n",
       "      <td>6.468091</td>\n",
       "      <td>3.838500</td>\n",
       "      <td>4.145614</td>\n",
       "      <td>1.747659</td>\n",
       "      <td>4.152091</td>\n",
       "      <td>...</td>\n",
       "      <td>56.073432</td>\n",
       "      <td>428.064205</td>\n",
       "      <td>5.310705</td>\n",
       "      <td>11.464273</td>\n",
       "      <td>3.527250</td>\n",
       "      <td>3.786523</td>\n",
       "      <td>3.907591</td>\n",
       "      <td>4.692114</td>\n",
       "      <td>12.210477</td>\n",
       "      <td>3.712727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004669</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.898068</td>\n",
       "      <td>5.377864</td>\n",
       "      <td>2.946705</td>\n",
       "      <td>3.163330</td>\n",
       "      <td>1.896011</td>\n",
       "      <td>1.022955</td>\n",
       "      <td>0.172182</td>\n",
       "      <td>...</td>\n",
       "      <td>61.539432</td>\n",
       "      <td>428.064205</td>\n",
       "      <td>4.940864</td>\n",
       "      <td>11.965091</td>\n",
       "      <td>5.069795</td>\n",
       "      <td>4.555818</td>\n",
       "      <td>1.995182</td>\n",
       "      <td>6.432477</td>\n",
       "      <td>11.655318</td>\n",
       "      <td>4.331432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004670</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.305909</td>\n",
       "      <td>7.954511</td>\n",
       "      <td>4.333091</td>\n",
       "      <td>3.442909</td>\n",
       "      <td>2.423591</td>\n",
       "      <td>1.638136</td>\n",
       "      <td>3.971659</td>\n",
       "      <td>...</td>\n",
       "      <td>63.646295</td>\n",
       "      <td>462.197295</td>\n",
       "      <td>6.389000</td>\n",
       "      <td>15.515795</td>\n",
       "      <td>4.944500</td>\n",
       "      <td>4.093750</td>\n",
       "      <td>2.188148</td>\n",
       "      <td>7.729409</td>\n",
       "      <td>16.610136</td>\n",
       "      <td>9.654886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004671</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.175523</td>\n",
       "      <td>6.139898</td>\n",
       "      <td>7.206341</td>\n",
       "      <td>2.149045</td>\n",
       "      <td>5.279455</td>\n",
       "      <td>0.992409</td>\n",
       "      <td>4.651966</td>\n",
       "      <td>...</td>\n",
       "      <td>66.292295</td>\n",
       "      <td>377.576568</td>\n",
       "      <td>6.499659</td>\n",
       "      <td>12.374955</td>\n",
       "      <td>1.602136</td>\n",
       "      <td>5.752091</td>\n",
       "      <td>2.603136</td>\n",
       "      <td>2.183420</td>\n",
       "      <td>12.136659</td>\n",
       "      <td>7.898068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004672</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.369318</td>\n",
       "      <td>5.205511</td>\n",
       "      <td>5.377205</td>\n",
       "      <td>2.704318</td>\n",
       "      <td>4.781898</td>\n",
       "      <td>2.732523</td>\n",
       "      <td>5.757386</td>\n",
       "      <td>...</td>\n",
       "      <td>56.073432</td>\n",
       "      <td>390.410159</td>\n",
       "      <td>5.003727</td>\n",
       "      <td>13.899705</td>\n",
       "      <td>1.856886</td>\n",
       "      <td>5.537568</td>\n",
       "      <td>1.267636</td>\n",
       "      <td>2.900750</td>\n",
       "      <td>10.118227</td>\n",
       "      <td>2.878568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004673</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.710773</td>\n",
       "      <td>6.203642</td>\n",
       "      <td>7.865614</td>\n",
       "      <td>5.034500</td>\n",
       "      <td>3.488557</td>\n",
       "      <td>2.280705</td>\n",
       "      <td>5.586114</td>\n",
       "      <td>...</td>\n",
       "      <td>61.539432</td>\n",
       "      <td>523.958114</td>\n",
       "      <td>5.547136</td>\n",
       "      <td>8.514114</td>\n",
       "      <td>3.290977</td>\n",
       "      <td>3.521932</td>\n",
       "      <td>1.715341</td>\n",
       "      <td>3.434205</td>\n",
       "      <td>15.863295</td>\n",
       "      <td>3.013386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004674</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.999045</td>\n",
       "      <td>7.332875</td>\n",
       "      <td>5.283455</td>\n",
       "      <td>3.070500</td>\n",
       "      <td>2.289409</td>\n",
       "      <td>2.414773</td>\n",
       "      <td>5.789523</td>\n",
       "      <td>...</td>\n",
       "      <td>50.676614</td>\n",
       "      <td>462.197295</td>\n",
       "      <td>6.992864</td>\n",
       "      <td>12.374955</td>\n",
       "      <td>1.672602</td>\n",
       "      <td>6.953432</td>\n",
       "      <td>1.177045</td>\n",
       "      <td>2.910886</td>\n",
       "      <td>17.465977</td>\n",
       "      <td>5.680886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004675</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.756705</td>\n",
       "      <td>5.051580</td>\n",
       "      <td>5.647227</td>\n",
       "      <td>1.114784</td>\n",
       "      <td>1.786250</td>\n",
       "      <td>2.316716</td>\n",
       "      <td>2.561864</td>\n",
       "      <td>...</td>\n",
       "      <td>51.473727</td>\n",
       "      <td>428.064205</td>\n",
       "      <td>6.279545</td>\n",
       "      <td>9.181409</td>\n",
       "      <td>2.564091</td>\n",
       "      <td>4.203136</td>\n",
       "      <td>1.114784</td>\n",
       "      <td>4.711818</td>\n",
       "      <td>13.583841</td>\n",
       "      <td>6.819295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004676</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.301864</td>\n",
       "      <td>5.115739</td>\n",
       "      <td>4.661909</td>\n",
       "      <td>2.160955</td>\n",
       "      <td>2.859659</td>\n",
       "      <td>2.338750</td>\n",
       "      <td>2.921341</td>\n",
       "      <td>...</td>\n",
       "      <td>50.423182</td>\n",
       "      <td>363.600432</td>\n",
       "      <td>4.921318</td>\n",
       "      <td>9.490727</td>\n",
       "      <td>1.947432</td>\n",
       "      <td>3.576977</td>\n",
       "      <td>3.020795</td>\n",
       "      <td>4.624545</td>\n",
       "      <td>11.564364</td>\n",
       "      <td>8.600409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004677</th>\n",
       "      <td>1.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.611000</td>\n",
       "      <td>5.730193</td>\n",
       "      <td>6.440432</td>\n",
       "      <td>0.912182</td>\n",
       "      <td>5.733977</td>\n",
       "      <td>3.386239</td>\n",
       "      <td>4.723693</td>\n",
       "      <td>...</td>\n",
       "      <td>49.512591</td>\n",
       "      <td>198.865909</td>\n",
       "      <td>6.884182</td>\n",
       "      <td>7.446591</td>\n",
       "      <td>2.640432</td>\n",
       "      <td>3.122591</td>\n",
       "      <td>2.050455</td>\n",
       "      <td>6.598591</td>\n",
       "      <td>20.016341</td>\n",
       "      <td>4.099205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004678</th>\n",
       "      <td>1.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.854136</td>\n",
       "      <td>6.067318</td>\n",
       "      <td>5.702545</td>\n",
       "      <td>2.234193</td>\n",
       "      <td>5.506409</td>\n",
       "      <td>1.394568</td>\n",
       "      <td>4.445000</td>\n",
       "      <td>...</td>\n",
       "      <td>46.477955</td>\n",
       "      <td>232.556864</td>\n",
       "      <td>7.044659</td>\n",
       "      <td>7.344795</td>\n",
       "      <td>1.793159</td>\n",
       "      <td>3.755886</td>\n",
       "      <td>3.261386</td>\n",
       "      <td>6.476136</td>\n",
       "      <td>19.825409</td>\n",
       "      <td>5.032182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004679</th>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.810523</td>\n",
       "      <td>4.527466</td>\n",
       "      <td>3.482795</td>\n",
       "      <td>6.671614</td>\n",
       "      <td>3.380341</td>\n",
       "      <td>2.722102</td>\n",
       "      <td>4.433023</td>\n",
       "      <td>...</td>\n",
       "      <td>44.692932</td>\n",
       "      <td>351.976114</td>\n",
       "      <td>4.890318</td>\n",
       "      <td>5.949545</td>\n",
       "      <td>5.182455</td>\n",
       "      <td>3.418818</td>\n",
       "      <td>2.373205</td>\n",
       "      <td>5.619250</td>\n",
       "      <td>22.588023</td>\n",
       "      <td>2.753659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004680</th>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.465023</td>\n",
       "      <td>5.047153</td>\n",
       "      <td>3.427091</td>\n",
       "      <td>5.208977</td>\n",
       "      <td>3.434205</td>\n",
       "      <td>2.594705</td>\n",
       "      <td>4.952057</td>\n",
       "      <td>...</td>\n",
       "      <td>47.485227</td>\n",
       "      <td>340.568750</td>\n",
       "      <td>5.331045</td>\n",
       "      <td>6.716409</td>\n",
       "      <td>4.498682</td>\n",
       "      <td>3.543114</td>\n",
       "      <td>3.004705</td>\n",
       "      <td>5.624705</td>\n",
       "      <td>24.326614</td>\n",
       "      <td>4.470068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004681</th>\n",
       "      <td>1.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.938318</td>\n",
       "      <td>6.115670</td>\n",
       "      <td>6.920977</td>\n",
       "      <td>1.309045</td>\n",
       "      <td>3.399432</td>\n",
       "      <td>2.757364</td>\n",
       "      <td>4.769750</td>\n",
       "      <td>...</td>\n",
       "      <td>53.607227</td>\n",
       "      <td>308.650705</td>\n",
       "      <td>5.776841</td>\n",
       "      <td>6.454250</td>\n",
       "      <td>2.595659</td>\n",
       "      <td>4.222091</td>\n",
       "      <td>2.539682</td>\n",
       "      <td>5.949545</td>\n",
       "      <td>17.372409</td>\n",
       "      <td>4.407591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004682</th>\n",
       "      <td>1.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.733841</td>\n",
       "      <td>5.925716</td>\n",
       "      <td>6.515409</td>\n",
       "      <td>3.503614</td>\n",
       "      <td>3.492955</td>\n",
       "      <td>2.361205</td>\n",
       "      <td>5.351864</td>\n",
       "      <td>...</td>\n",
       "      <td>47.707295</td>\n",
       "      <td>266.615704</td>\n",
       "      <td>5.200114</td>\n",
       "      <td>6.321307</td>\n",
       "      <td>0.559159</td>\n",
       "      <td>3.948386</td>\n",
       "      <td>2.653182</td>\n",
       "      <td>7.113432</td>\n",
       "      <td>15.710773</td>\n",
       "      <td>5.638795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004683</th>\n",
       "      <td>1.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.116273</td>\n",
       "      <td>7.523545</td>\n",
       "      <td>7.524295</td>\n",
       "      <td>5.547136</td>\n",
       "      <td>3.569807</td>\n",
       "      <td>5.840773</td>\n",
       "      <td>7.739000</td>\n",
       "      <td>...</td>\n",
       "      <td>56.879750</td>\n",
       "      <td>46.142227</td>\n",
       "      <td>6.598591</td>\n",
       "      <td>5.343114</td>\n",
       "      <td>2.866432</td>\n",
       "      <td>3.858955</td>\n",
       "      <td>2.343091</td>\n",
       "      <td>0.838932</td>\n",
       "      <td>21.436455</td>\n",
       "      <td>3.431318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004684</th>\n",
       "      <td>1.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.374955</td>\n",
       "      <td>8.381591</td>\n",
       "      <td>6.880375</td>\n",
       "      <td>6.660886</td>\n",
       "      <td>4.829159</td>\n",
       "      <td>6.238136</td>\n",
       "      <td>6.867818</td>\n",
       "      <td>...</td>\n",
       "      <td>47.485227</td>\n",
       "      <td>71.752114</td>\n",
       "      <td>6.305159</td>\n",
       "      <td>5.871818</td>\n",
       "      <td>1.864318</td>\n",
       "      <td>3.884727</td>\n",
       "      <td>2.691670</td>\n",
       "      <td>1.217398</td>\n",
       "      <td>21.803500</td>\n",
       "      <td>3.431318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004685</th>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.169455</td>\n",
       "      <td>3.822807</td>\n",
       "      <td>6.037114</td>\n",
       "      <td>1.916841</td>\n",
       "      <td>8.851568</td>\n",
       "      <td>3.372273</td>\n",
       "      <td>6.394932</td>\n",
       "      <td>...</td>\n",
       "      <td>59.032955</td>\n",
       "      <td>284.082750</td>\n",
       "      <td>3.879136</td>\n",
       "      <td>3.814852</td>\n",
       "      <td>1.916841</td>\n",
       "      <td>1.916841</td>\n",
       "      <td>3.409023</td>\n",
       "      <td>3.287386</td>\n",
       "      <td>25.236659</td>\n",
       "      <td>3.437091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004686</th>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.752091</td>\n",
       "      <td>4.408455</td>\n",
       "      <td>6.260898</td>\n",
       "      <td>3.121273</td>\n",
       "      <td>10.733841</td>\n",
       "      <td>2.601659</td>\n",
       "      <td>5.003727</td>\n",
       "      <td>...</td>\n",
       "      <td>36.598795</td>\n",
       "      <td>363.600432</td>\n",
       "      <td>4.856523</td>\n",
       "      <td>2.945727</td>\n",
       "      <td>1.908386</td>\n",
       "      <td>3.001250</td>\n",
       "      <td>3.390432</td>\n",
       "      <td>2.748455</td>\n",
       "      <td>25.599886</td>\n",
       "      <td>4.921318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004687</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.957818</td>\n",
       "      <td>4.428437</td>\n",
       "      <td>7.904307</td>\n",
       "      <td>2.993955</td>\n",
       "      <td>3.976977</td>\n",
       "      <td>1.877750</td>\n",
       "      <td>3.887000</td>\n",
       "      <td>...</td>\n",
       "      <td>64.459364</td>\n",
       "      <td>314.541023</td>\n",
       "      <td>5.909795</td>\n",
       "      <td>8.444795</td>\n",
       "      <td>3.789227</td>\n",
       "      <td>2.840909</td>\n",
       "      <td>2.460250</td>\n",
       "      <td>6.571159</td>\n",
       "      <td>27.583023</td>\n",
       "      <td>7.235409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004688</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.022886</td>\n",
       "      <td>4.148011</td>\n",
       "      <td>8.421682</td>\n",
       "      <td>3.789227</td>\n",
       "      <td>4.533591</td>\n",
       "      <td>2.745636</td>\n",
       "      <td>2.998864</td>\n",
       "      <td>...</td>\n",
       "      <td>78.123091</td>\n",
       "      <td>322.905432</td>\n",
       "      <td>5.161818</td>\n",
       "      <td>5.888682</td>\n",
       "      <td>3.749136</td>\n",
       "      <td>2.752148</td>\n",
       "      <td>3.605705</td>\n",
       "      <td>5.760727</td>\n",
       "      <td>23.726841</td>\n",
       "      <td>4.132818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004689</th>\n",
       "      <td>1.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.696045</td>\n",
       "      <td>5.583023</td>\n",
       "      <td>6.651886</td>\n",
       "      <td>4.423670</td>\n",
       "      <td>1.439886</td>\n",
       "      <td>2.619080</td>\n",
       "      <td>3.094227</td>\n",
       "      <td>...</td>\n",
       "      <td>68.054591</td>\n",
       "      <td>261.229318</td>\n",
       "      <td>7.409523</td>\n",
       "      <td>6.454250</td>\n",
       "      <td>1.426205</td>\n",
       "      <td>3.460523</td>\n",
       "      <td>1.370273</td>\n",
       "      <td>3.263795</td>\n",
       "      <td>15.005955</td>\n",
       "      <td>4.706273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004690</th>\n",
       "      <td>1.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.371818</td>\n",
       "      <td>4.967341</td>\n",
       "      <td>6.834841</td>\n",
       "      <td>4.664216</td>\n",
       "      <td>2.184136</td>\n",
       "      <td>3.421182</td>\n",
       "      <td>3.431318</td>\n",
       "      <td>...</td>\n",
       "      <td>54.949886</td>\n",
       "      <td>322.905432</td>\n",
       "      <td>9.305432</td>\n",
       "      <td>6.530250</td>\n",
       "      <td>1.702091</td>\n",
       "      <td>3.260307</td>\n",
       "      <td>2.416659</td>\n",
       "      <td>3.665227</td>\n",
       "      <td>14.472636</td>\n",
       "      <td>7.139318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004691</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.591148</td>\n",
       "      <td>5.021307</td>\n",
       "      <td>5.958523</td>\n",
       "      <td>5.718386</td>\n",
       "      <td>1.479148</td>\n",
       "      <td>1.479148</td>\n",
       "      <td>6.908250</td>\n",
       "      <td>...</td>\n",
       "      <td>56.073432</td>\n",
       "      <td>340.568750</td>\n",
       "      <td>4.990045</td>\n",
       "      <td>6.377568</td>\n",
       "      <td>3.127250</td>\n",
       "      <td>2.971625</td>\n",
       "      <td>1.479148</td>\n",
       "      <td>4.000250</td>\n",
       "      <td>22.955568</td>\n",
       "      <td>3.619250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004692</th>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.100955</td>\n",
       "      <td>5.170159</td>\n",
       "      <td>6.061705</td>\n",
       "      <td>6.238136</td>\n",
       "      <td>1.606227</td>\n",
       "      <td>2.093614</td>\n",
       "      <td>6.567659</td>\n",
       "      <td>...</td>\n",
       "      <td>44.997636</td>\n",
       "      <td>340.568750</td>\n",
       "      <td>5.551636</td>\n",
       "      <td>7.353205</td>\n",
       "      <td>2.674750</td>\n",
       "      <td>3.644750</td>\n",
       "      <td>2.442682</td>\n",
       "      <td>4.701227</td>\n",
       "      <td>21.140386</td>\n",
       "      <td>4.172500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004693</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.280591</td>\n",
       "      <td>4.953369</td>\n",
       "      <td>6.756705</td>\n",
       "      <td>4.003159</td>\n",
       "      <td>4.489114</td>\n",
       "      <td>2.275125</td>\n",
       "      <td>3.553477</td>\n",
       "      <td>...</td>\n",
       "      <td>53.607227</td>\n",
       "      <td>377.576568</td>\n",
       "      <td>6.451545</td>\n",
       "      <td>6.786727</td>\n",
       "      <td>2.580295</td>\n",
       "      <td>2.470284</td>\n",
       "      <td>2.807750</td>\n",
       "      <td>7.104273</td>\n",
       "      <td>17.286795</td>\n",
       "      <td>2.942284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004694</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.148239</td>\n",
       "      <td>5.731830</td>\n",
       "      <td>6.643341</td>\n",
       "      <td>5.027864</td>\n",
       "      <td>4.833023</td>\n",
       "      <td>1.753545</td>\n",
       "      <td>4.123159</td>\n",
       "      <td>...</td>\n",
       "      <td>53.024932</td>\n",
       "      <td>390.410159</td>\n",
       "      <td>5.942318</td>\n",
       "      <td>7.446591</td>\n",
       "      <td>3.401432</td>\n",
       "      <td>3.244091</td>\n",
       "      <td>4.364205</td>\n",
       "      <td>7.253159</td>\n",
       "      <td>16.766295</td>\n",
       "      <td>5.166523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004695</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.108727</td>\n",
       "      <td>5.867227</td>\n",
       "      <td>8.088909</td>\n",
       "      <td>2.009568</td>\n",
       "      <td>5.506409</td>\n",
       "      <td>2.614023</td>\n",
       "      <td>4.463545</td>\n",
       "      <td>...</td>\n",
       "      <td>51.473727</td>\n",
       "      <td>299.711318</td>\n",
       "      <td>6.432477</td>\n",
       "      <td>5.299955</td>\n",
       "      <td>2.515364</td>\n",
       "      <td>4.407591</td>\n",
       "      <td>3.128023</td>\n",
       "      <td>8.872409</td>\n",
       "      <td>19.003841</td>\n",
       "      <td>7.104273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004696</th>\n",
       "      <td>1.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.756568</td>\n",
       "      <td>5.509068</td>\n",
       "      <td>6.793795</td>\n",
       "      <td>0.258432</td>\n",
       "      <td>3.971659</td>\n",
       "      <td>3.314443</td>\n",
       "      <td>6.284818</td>\n",
       "      <td>...</td>\n",
       "      <td>51.821636</td>\n",
       "      <td>405.842364</td>\n",
       "      <td>6.293795</td>\n",
       "      <td>4.203136</td>\n",
       "      <td>2.873136</td>\n",
       "      <td>3.291898</td>\n",
       "      <td>3.073705</td>\n",
       "      <td>9.100227</td>\n",
       "      <td>20.658341</td>\n",
       "      <td>6.187409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004697</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.489977</td>\n",
       "      <td>6.256591</td>\n",
       "      <td>6.476136</td>\n",
       "      <td>2.962682</td>\n",
       "      <td>3.641545</td>\n",
       "      <td>2.475182</td>\n",
       "      <td>4.445000</td>\n",
       "      <td>...</td>\n",
       "      <td>43.263341</td>\n",
       "      <td>331.532841</td>\n",
       "      <td>6.104318</td>\n",
       "      <td>6.166136</td>\n",
       "      <td>2.608477</td>\n",
       "      <td>3.743045</td>\n",
       "      <td>2.769273</td>\n",
       "      <td>4.407591</td>\n",
       "      <td>17.032136</td>\n",
       "      <td>5.557045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1004698</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.909795</td>\n",
       "      <td>4.950045</td>\n",
       "      <td>7.770227</td>\n",
       "      <td>2.970159</td>\n",
       "      <td>3.018000</td>\n",
       "      <td>3.885864</td>\n",
       "      <td>3.537773</td>\n",
       "      <td>...</td>\n",
       "      <td>39.078068</td>\n",
       "      <td>314.541023</td>\n",
       "      <td>9.756568</td>\n",
       "      <td>4.872477</td>\n",
       "      <td>1.887273</td>\n",
       "      <td>2.431773</td>\n",
       "      <td>2.724955</td>\n",
       "      <td>3.595500</td>\n",
       "      <td>19.342568</td>\n",
       "      <td>9.539614</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44 rows × 2276 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Kidney Chromophobe   Age  Gender      A2ML1    AARSD1      AATF  \\\n",
       "GSM1004655                 1.0  78.0     1.0  11.611000  5.463477  5.866136   \n",
       "GSM1004656                 1.0  78.0     1.0  12.398614  7.561534  6.779500   \n",
       "GSM1004657                 1.0  53.0     0.0   8.044136  4.762284  7.152886   \n",
       "GSM1004658                 1.0  53.0     0.0   8.556591  4.645347  7.174977   \n",
       "GSM1004659                 1.0  71.0     1.0   7.704807  8.187159  5.923523   \n",
       "GSM1004660                 1.0  71.0     1.0   6.522909  4.843648  7.814091   \n",
       "GSM1004661                 1.0  39.0     0.0   9.999795  5.484097  5.647227   \n",
       "GSM1004662                 1.0  39.0     0.0   9.772500  6.125852  5.936205   \n",
       "GSM1004663                 1.0  34.0     1.0   7.132091  4.627341  5.132614   \n",
       "GSM1004664                 1.0  34.0     1.0   5.202318  4.554318  8.291773   \n",
       "GSM1004665                 1.0  51.0     0.0   7.729409  5.508659  6.432477   \n",
       "GSM1004666                 1.0  51.0     0.0   6.749909  4.967398  5.235159   \n",
       "GSM1004667                 1.0  75.0     0.0   9.229159  4.116148  8.291773   \n",
       "GSM1004668                 1.0  75.0     0.0   9.699705  5.179080  6.468091   \n",
       "GSM1004669                 1.0  40.0     1.0   7.898068  5.377864  2.946705   \n",
       "GSM1004670                 1.0  40.0     1.0   7.305909  7.954511  4.333091   \n",
       "GSM1004671                 1.0  51.0     0.0  11.175523  6.139898  7.206341   \n",
       "GSM1004672                 1.0  51.0     0.0  11.369318  5.205511  5.377205   \n",
       "GSM1004673                 1.0  51.0     1.0   6.710773  6.203642  7.865614   \n",
       "GSM1004674                 1.0  51.0     1.0   7.999045  7.332875  5.283455   \n",
       "GSM1004675                 1.0  50.0     0.0   6.756705  5.051580  5.647227   \n",
       "GSM1004676                 1.0  50.0     0.0   6.301864  5.115739  4.661909   \n",
       "GSM1004677                 1.0  78.0     1.0  11.611000  5.730193  6.440432   \n",
       "GSM1004678                 1.0  78.0     1.0  10.854136  6.067318  5.702545   \n",
       "GSM1004679                 1.0  53.0     0.0   6.810523  4.527466  3.482795   \n",
       "GSM1004680                 1.0  53.0     0.0   7.465023  5.047153  3.427091   \n",
       "GSM1004681                 1.0  71.0     1.0  11.938318  6.115670  6.920977   \n",
       "GSM1004682                 1.0  71.0     1.0  10.733841  5.925716  6.515409   \n",
       "GSM1004683                 1.0  39.0     0.0  12.116273  7.523545  7.524295   \n",
       "GSM1004684                 1.0  39.0     0.0  12.374955  8.381591  6.880375   \n",
       "GSM1004685                 1.0  34.0     1.0   9.169455  3.822807  6.037114   \n",
       "GSM1004686                 1.0  34.0     1.0   5.752091  4.408455  6.260898   \n",
       "GSM1004687                 1.0  51.0     0.0   7.957818  4.428437  7.904307   \n",
       "GSM1004688                 1.0  51.0     0.0   6.022886  4.148011  8.421682   \n",
       "GSM1004689                 1.0  75.0     0.0   8.696045  5.583023  6.651886   \n",
       "GSM1004690                 1.0  75.0     0.0   7.371818  4.967341  6.834841   \n",
       "GSM1004691                 1.0  40.0     1.0   8.591148  5.021307  5.958523   \n",
       "GSM1004692                 1.0  40.0     1.0   8.100955  5.170159  6.061705   \n",
       "GSM1004693                 1.0  51.0     0.0  11.280591  4.953369  6.756705   \n",
       "GSM1004694                 1.0  51.0     0.0  11.148239  5.731830  6.643341   \n",
       "GSM1004695                 1.0  51.0     1.0  11.108727  5.867227  8.088909   \n",
       "GSM1004696                 1.0  51.0     1.0   9.756568  5.509068  6.793795   \n",
       "GSM1004697                 1.0  50.0     0.0   8.489977  6.256591  6.476136   \n",
       "GSM1004698                 1.0  50.0     0.0   5.909795  4.950045  7.770227   \n",
       "\n",
       "               ABCA3     ABCB10     ABCB6     ABCB7  ...     ZNF841  \\\n",
       "GSM1004655  2.304568   3.128023  2.288523  3.816227  ...  52.649727   \n",
       "GSM1004656  2.531636   4.258580  1.857784  5.336545  ...  53.024932   \n",
       "GSM1004657  4.169409   5.121580  2.802159  5.929659  ...  56.879750   \n",
       "GSM1004658  3.884727   5.776841  3.129000  5.627193  ...  53.339636   \n",
       "GSM1004659  0.992409   5.195102  1.369386  4.128000  ...  61.150727   \n",
       "GSM1004660  2.001591   6.675045  3.592977  4.304909  ...  57.500205   \n",
       "GSM1004661  2.025750   5.629682  4.240318  4.577591  ...  62.049091   \n",
       "GSM1004662  4.709955   5.837705  3.957761  6.022886  ...  52.649727   \n",
       "GSM1004663  2.187364   6.154091  1.673659  3.484932  ...  61.539432   \n",
       "GSM1004664  2.650977   6.749909  3.503614  2.606432  ...  53.883227   \n",
       "GSM1004665  2.658307   3.019966  0.893205  3.519932  ...  65.250318   \n",
       "GSM1004666  4.222091   2.882489  1.199545  4.421409  ...  68.054591   \n",
       "GSM1004667  2.534909   4.392636  2.126216  2.981818  ...  52.142364   \n",
       "GSM1004668  3.838500   4.145614  1.747659  4.152091  ...  56.073432   \n",
       "GSM1004669  3.163330   1.896011  1.022955  0.172182  ...  61.539432   \n",
       "GSM1004670  3.442909   2.423591  1.638136  3.971659  ...  63.646295   \n",
       "GSM1004671  2.149045   5.279455  0.992409  4.651966  ...  66.292295   \n",
       "GSM1004672  2.704318   4.781898  2.732523  5.757386  ...  56.073432   \n",
       "GSM1004673  5.034500   3.488557  2.280705  5.586114  ...  61.539432   \n",
       "GSM1004674  3.070500   2.289409  2.414773  5.789523  ...  50.676614   \n",
       "GSM1004675  1.114784   1.786250  2.316716  2.561864  ...  51.473727   \n",
       "GSM1004676  2.160955   2.859659  2.338750  2.921341  ...  50.423182   \n",
       "GSM1004677  0.912182   5.733977  3.386239  4.723693  ...  49.512591   \n",
       "GSM1004678  2.234193   5.506409  1.394568  4.445000  ...  46.477955   \n",
       "GSM1004679  6.671614   3.380341  2.722102  4.433023  ...  44.692932   \n",
       "GSM1004680  5.208977   3.434205  2.594705  4.952057  ...  47.485227   \n",
       "GSM1004681  1.309045   3.399432  2.757364  4.769750  ...  53.607227   \n",
       "GSM1004682  3.503614   3.492955  2.361205  5.351864  ...  47.707295   \n",
       "GSM1004683  5.547136   3.569807  5.840773  7.739000  ...  56.879750   \n",
       "GSM1004684  6.660886   4.829159  6.238136  6.867818  ...  47.485227   \n",
       "GSM1004685  1.916841   8.851568  3.372273  6.394932  ...  59.032955   \n",
       "GSM1004686  3.121273  10.733841  2.601659  5.003727  ...  36.598795   \n",
       "GSM1004687  2.993955   3.976977  1.877750  3.887000  ...  64.459364   \n",
       "GSM1004688  3.789227   4.533591  2.745636  2.998864  ...  78.123091   \n",
       "GSM1004689  4.423670   1.439886  2.619080  3.094227  ...  68.054591   \n",
       "GSM1004690  4.664216   2.184136  3.421182  3.431318  ...  54.949886   \n",
       "GSM1004691  5.718386   1.479148  1.479148  6.908250  ...  56.073432   \n",
       "GSM1004692  6.238136   1.606227  2.093614  6.567659  ...  44.997636   \n",
       "GSM1004693  4.003159   4.489114  2.275125  3.553477  ...  53.607227   \n",
       "GSM1004694  5.027864   4.833023  1.753545  4.123159  ...  53.024932   \n",
       "GSM1004695  2.009568   5.506409  2.614023  4.463545  ...  51.473727   \n",
       "GSM1004696  0.258432   3.971659  3.314443  6.284818  ...  51.821636   \n",
       "GSM1004697  2.962682   3.641545  2.475182  4.445000  ...  43.263341   \n",
       "GSM1004698  2.970159   3.018000  3.885864  3.537773  ...  39.078068   \n",
       "\n",
       "                ZNHIT3    ZRANB3    ZSCAN18      ZW10      ZXDC     ZZEF1  \\\n",
       "GSM1004655  428.064205  5.563250   8.566886  2.134705  2.842023  1.811693   \n",
       "GSM1004656  428.064205  4.555818  10.590000  1.646068  5.670727  2.781568   \n",
       "GSM1004657  492.838000  6.010773   7.792432  3.739795  3.114318  2.068795   \n",
       "GSM1004658  492.838000  6.126432   8.121682  3.108023  3.013386  2.061159   \n",
       "GSM1004659  559.608205  4.766364   8.009045  1.378205  7.037091  2.292614   \n",
       "GSM1004660  559.608205  4.132818   7.491523  2.551955  2.927023  2.279068   \n",
       "GSM1004661  110.698523  4.736011   8.182250  2.531636  4.182432  2.309682   \n",
       "GSM1004662  175.147705  4.940864   8.077614  2.713705  4.116295  3.003500   \n",
       "GSM1004663  363.600432  6.819295   6.194659  1.844273  3.781341  3.627909   \n",
       "GSM1004664  351.976114  7.990250   4.994000  1.340500  4.459932  2.482614   \n",
       "GSM1004665  405.842364  9.504591  10.311523  3.673341  4.847295  2.849705   \n",
       "GSM1004666  462.197295  4.605477  14.361295  3.524841  3.339045  3.824807   \n",
       "GSM1004667  405.842364  7.095455   6.424318  2.035216  2.671386  3.575364   \n",
       "GSM1004668  428.064205  5.310705  11.464273  3.527250  3.786523  3.907591   \n",
       "GSM1004669  428.064205  4.940864  11.965091  5.069795  4.555818  1.995182   \n",
       "GSM1004670  462.197295  6.389000  15.515795  4.944500  4.093750  2.188148   \n",
       "GSM1004671  377.576568  6.499659  12.374955  1.602136  5.752091  2.603136   \n",
       "GSM1004672  390.410159  5.003727  13.899705  1.856886  5.537568  1.267636   \n",
       "GSM1004673  523.958114  5.547136   8.514114  3.290977  3.521932  1.715341   \n",
       "GSM1004674  462.197295  6.992864  12.374955  1.672602  6.953432  1.177045   \n",
       "GSM1004675  428.064205  6.279545   9.181409  2.564091  4.203136  1.114784   \n",
       "GSM1004676  363.600432  4.921318   9.490727  1.947432  3.576977  3.020795   \n",
       "GSM1004677  198.865909  6.884182   7.446591  2.640432  3.122591  2.050455   \n",
       "GSM1004678  232.556864  7.044659   7.344795  1.793159  3.755886  3.261386   \n",
       "GSM1004679  351.976114  4.890318   5.949545  5.182455  3.418818  2.373205   \n",
       "GSM1004680  340.568750  5.331045   6.716409  4.498682  3.543114  3.004705   \n",
       "GSM1004681  308.650705  5.776841   6.454250  2.595659  4.222091  2.539682   \n",
       "GSM1004682  266.615704  5.200114   6.321307  0.559159  3.948386  2.653182   \n",
       "GSM1004683   46.142227  6.598591   5.343114  2.866432  3.858955  2.343091   \n",
       "GSM1004684   71.752114  6.305159   5.871818  1.864318  3.884727  2.691670   \n",
       "GSM1004685  284.082750  3.879136   3.814852  1.916841  1.916841  3.409023   \n",
       "GSM1004686  363.600432  4.856523   2.945727  1.908386  3.001250  3.390432   \n",
       "GSM1004687  314.541023  5.909795   8.444795  3.789227  2.840909  2.460250   \n",
       "GSM1004688  322.905432  5.161818   5.888682  3.749136  2.752148  3.605705   \n",
       "GSM1004689  261.229318  7.409523   6.454250  1.426205  3.460523  1.370273   \n",
       "GSM1004690  322.905432  9.305432   6.530250  1.702091  3.260307  2.416659   \n",
       "GSM1004691  340.568750  4.990045   6.377568  3.127250  2.971625  1.479148   \n",
       "GSM1004692  340.568750  5.551636   7.353205  2.674750  3.644750  2.442682   \n",
       "GSM1004693  377.576568  6.451545   6.786727  2.580295  2.470284  2.807750   \n",
       "GSM1004694  390.410159  5.942318   7.446591  3.401432  3.244091  4.364205   \n",
       "GSM1004695  299.711318  6.432477   5.299955  2.515364  4.407591  3.128023   \n",
       "GSM1004696  405.842364  6.293795   4.203136  2.873136  3.291898  3.073705   \n",
       "GSM1004697  331.532841  6.104318   6.166136  2.608477  3.743045  2.769273   \n",
       "GSM1004698  314.541023  9.756568   4.872477  1.887273  2.431773  2.724955   \n",
       "\n",
       "               ARL5A       NONO   PRKRIP1  \n",
       "GSM1004655  5.899591  18.844500  3.841659  \n",
       "GSM1004656  5.942318  18.554023  8.121682  \n",
       "GSM1004657  5.942318  21.731909  3.080023  \n",
       "GSM1004658  6.585159  21.803500  3.413045  \n",
       "GSM1004659  6.675045  14.853091  8.044136  \n",
       "GSM1004660  6.543614  15.429045  5.078136  \n",
       "GSM1004661  1.284795  18.751341  3.000057  \n",
       "GSM1004662  2.230034  17.925205  4.526136  \n",
       "GSM1004663  3.418818  21.613500  3.093068  \n",
       "GSM1004664  2.890795  21.613500  3.139205  \n",
       "GSM1004665  5.713614  22.483886  3.743045  \n",
       "GSM1004666  8.105477  16.566273  9.601432  \n",
       "GSM1004667  4.605477  14.588636  3.259227  \n",
       "GSM1004668  4.692114  12.210477  3.712727  \n",
       "GSM1004669  6.432477  11.655318  4.331432  \n",
       "GSM1004670  7.729409  16.610136  9.654886  \n",
       "GSM1004671  2.183420  12.136659  7.898068  \n",
       "GSM1004672  2.900750  10.118227  2.878568  \n",
       "GSM1004673  3.434205  15.863295  3.013386  \n",
       "GSM1004674  2.910886  17.465977  5.680886  \n",
       "GSM1004675  4.711818  13.583841  6.819295  \n",
       "GSM1004676  4.624545  11.564364  8.600409  \n",
       "GSM1004677  6.598591  20.016341  4.099205  \n",
       "GSM1004678  6.476136  19.825409  5.032182  \n",
       "GSM1004679  5.619250  22.588023  2.753659  \n",
       "GSM1004680  5.624705  24.326614  4.470068  \n",
       "GSM1004681  5.949545  17.372409  4.407591  \n",
       "GSM1004682  7.113432  15.710773  5.638795  \n",
       "GSM1004683  0.838932  21.436455  3.431318  \n",
       "GSM1004684  1.217398  21.803500  3.431318  \n",
       "GSM1004685  3.287386  25.236659  3.437091  \n",
       "GSM1004686  2.748455  25.599886  4.921318  \n",
       "GSM1004687  6.571159  27.583023  7.235409  \n",
       "GSM1004688  5.760727  23.726841  4.132818  \n",
       "GSM1004689  3.263795  15.005955  4.706273  \n",
       "GSM1004690  3.665227  14.472636  7.139318  \n",
       "GSM1004691  4.000250  22.955568  3.619250  \n",
       "GSM1004692  4.701227  21.140386  4.172500  \n",
       "GSM1004693  7.104273  17.286795  2.942284  \n",
       "GSM1004694  7.253159  16.766295  5.166523  \n",
       "GSM1004695  8.872409  19.003841  7.104273  \n",
       "GSM1004696  9.100227  20.658341  6.187409  \n",
       "GSM1004697  4.407591  17.032136  5.557045  \n",
       "GSM1004698  3.595500  19.342568  9.539614  \n",
       "\n",
       "[44 rows x 2276 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 44 samples.\n",
      "For the feature 'Kidney Chromophobe', the least common label is '1.0' with 44 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 40.0\n",
      "  50% (Median): 51.0\n",
      "  75%: 71.0\n",
      "Min: 34.0\n",
      "Max: 78.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 20 occurrences. This represents 45.45% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE19982\\\\GSE19982_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE19982\\\\GSE19982_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE19982\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE19949\\\\GSE19949_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE19949\\\\GSE19949-GPL3921_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious traiy convert\n",
    "cohort = accession_num = \"GSE19949\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE95425\\\\GSE95425_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE95425\\\\GSE95425_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished: Biased\n",
    "cohort = accession_num = \"GSE95425\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE8271\\\\GSE8271_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE8271\\\\GSE8271-GPL2004_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No gene mapping\n",
    "cohort = accession_num = \"GSE8271\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE144082\\\\GSE144082_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE144082\\\\GSE144082_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No gene mapping\n",
    "cohort = accession_num = \"GSE144082\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE11024\\\\GSE11024_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE11024\\\\GSE11024_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE11024\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE11447\\\\GSE11447_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE11447\\\\GSE11447_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No obvious traits\n",
    "cohort = accession_num = \"GSE11447\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE3\\\\GSE3_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE3\\\\GSE3-GPL10_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stop: No clinical data\n",
    "cohort = accession_num = \"GSE3\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE6280\\\\GSE6280_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE6280\\\\GSE6280-GPL96_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE6280\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE11151\\\\GSE11151_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE11151\\\\GSE11151_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE11151\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE4125\\\\GSE4125_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE4125\\\\GSE4125-GPL2649_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stopped: No obvious traits\n",
    "cohort = accession_num = \"GSE4125\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE40912\\\\GSE40912_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE40912\\\\GSE40912_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished: Biased\n",
    "cohort = accession_num = \"GSE40912\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE57162\\\\GSE57162_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE57162\\\\GSE57162_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE57162\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE40914\\\\GSE40914_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE40914\\\\GSE40914-GPL3985_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished: Biased\n",
    "cohort = accession_num = \"GSE40914\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE68606\\\\GSE68606_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Kidney-Chromophobe\\\\GSE68606\\\\GSE68606_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE68606\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepossessing\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Integrative genome-wide expression profiling identifies three distinct molecular subgroups of renal cell carcinoma with different patient outcome\"\n",
      "!Series_summary\t\"Background: Renal cell carcinoma (RCC) is characterized by a number of diverse molecular aberrations that differ among individuals. Recent approaches to molecularly classify RCC were based on clinical, pathological as well as on single molecular parameters. As a consequence, gene expression patterns reflecting the sum of genetic aberrations in individual tumors may not have been recognized. In an attempt to uncover such molecular features in RCC, we used a novel, unbiased and integrative approach.\"\n",
      "!Series_summary\t\"Methods: We integrated gene expression data from 97 primary RCCs of different pathologic parameters, 15 RCC metastases as well as 34 cancer cell lines for two-way nonsupervised hierarchical clustering using gene groups suggested by the PANTHER Classification System. We depicted the genomic landscape of the resulted tumor groups by means of Single Nuclear Polymorphism (SNP) technology. Finally, the achieved results were immunohistochemically analyzed using a tissue microarray (TMA) composed of 254 RCC. Results: We found robust, genome wide expression signatures, which split RCC into three distinct molecular subgroups. These groups remained stable even if randomly selected gene sets were clustered. Notably, the pattern obtained from RCC cell lines was clearly distinguishable from that of primary tumors. SNP array analysis demonstrated differing frequencies of chromosomal copy number alterations among RCC subgroups. TMA analysis with group-specific markers showed a prognostic significance of the different groups. Conclusion: We propose the existence of characteristic and histologically independent genome-wide expression outputs in RCC with potential biological and clinical relevance.\"\n",
      "!Series_overall_design\t\"Expression profiling by array, combined data analysis with genomic profiling data. Genomic DNA from renal cell was hybridized to renal cell carcinoma samples and matched normal kidney tissue biopsies, using the Affymetrix GenomewideSNP_6 platform. CEL files were processed using R, Bioconductor and software from the aroma.affymetrix project. Visualized Copy number profiles are accessible through the Progenetix site (www.progenetix.net). CN,raw.csv and segments.csv: Probes are mapped by their position in genome build 36 / HG18. Probes are ordered according to their linear position on the Golden Path.\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM498450</th>\n",
       "      <th>GSM498451</th>\n",
       "      <th>GSM498452</th>\n",
       "      <th>GSM498453</th>\n",
       "      <th>GSM498454</th>\n",
       "      <th>GSM498455</th>\n",
       "      <th>GSM498456</th>\n",
       "      <th>GSM498457</th>\n",
       "      <th>GSM498458</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM498587</th>\n",
       "      <th>GSM498588</th>\n",
       "      <th>GSM498589</th>\n",
       "      <th>GSM498590</th>\n",
       "      <th>GSM498591</th>\n",
       "      <th>GSM498592</th>\n",
       "      <th>GSM498593</th>\n",
       "      <th>GSM498594</th>\n",
       "      <th>GSM498595</th>\n",
       "      <th>GSM498596</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>grade: 2</td>\n",
       "      <td>grade: 2</td>\n",
       "      <td>grade: 2</td>\n",
       "      <td>grade: 2</td>\n",
       "      <td>grade: 1</td>\n",
       "      <td>grade: 2</td>\n",
       "      <td>grade: 1</td>\n",
       "      <td>grade: 2</td>\n",
       "      <td>grade: 3</td>\n",
       "      <td>...</td>\n",
       "      <td>cell line: SLR20</td>\n",
       "      <td>cell line: Caki-2</td>\n",
       "      <td>cell line: SLR21</td>\n",
       "      <td>cell line: KU19-20</td>\n",
       "      <td>cell line: SLR23</td>\n",
       "      <td>cell line: PC3hep27</td>\n",
       "      <td>cell line: PC3hep30</td>\n",
       "      <td>cell line: PC3vec1</td>\n",
       "      <td>cell line: PC3vec3</td>\n",
       "      <td>cell line: HK-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>stage: 2</td>\n",
       "      <td>stage: 2</td>\n",
       "      <td>stage: 2</td>\n",
       "      <td>stage: 1</td>\n",
       "      <td>stage: 2</td>\n",
       "      <td>stage: 2</td>\n",
       "      <td>stage: 1</td>\n",
       "      <td>stage: 2</td>\n",
       "      <td>stage: 3</td>\n",
       "      <td>...</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "      <td>grade: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>...</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "      <td>stage: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>icd-o 3 code: 8310/3</td>\n",
       "      <td>...</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "      <td>sample type: neoplasia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>icd-o 3 diagnosis text: clear cell renal cell ...</td>\n",
       "      <td>...</td>\n",
       "      <td>icd-o 3 code: 8312/3</td>\n",
       "      <td>icd-o 3 code: 8312/3</td>\n",
       "      <td>icd-o 3 code: 8312/3</td>\n",
       "      <td>icd-o 3 code: 8312/3</td>\n",
       "      <td>icd-o 3 code: 8312/3</td>\n",
       "      <td>icd-o 3 code: 8140/3</td>\n",
       "      <td>icd-o 3 code: 8140/3</td>\n",
       "      <td>icd-o 3 code: 8140/3</td>\n",
       "      <td>icd-o 3 code: 8140/3</td>\n",
       "      <td>icd-o 3 code: 8312/3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>organ site: kidney</td>\n",
       "      <td>...</td>\n",
       "      <td>icd-o 3 diagnosis text: renal cell carcinoma</td>\n",
       "      <td>icd-o 3 diagnosis text: renal cell carcinoma</td>\n",
       "      <td>icd-o 3 diagnosis text: renal cell carcinoma</td>\n",
       "      <td>icd-o 3 diagnosis text: renal cell carcinoma</td>\n",
       "      <td>icd-o 3 diagnosis text: renal cell carcinoma</td>\n",
       "      <td>icd-o 3 diagnosis text: adenocarcinoma, NOS</td>\n",
       "      <td>icd-o 3 diagnosis text: adenocarcinoma, NOS</td>\n",
       "      <td>icd-o 3 diagnosis text: adenocarcinoma, NOS</td>\n",
       "      <td>icd-o 3 diagnosis text: adenocarcinoma, NOS</td>\n",
       "      <td>icd-o 3 diagnosis text: renal cell carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: female</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>...</td>\n",
       "      <td>organ site: kidney [cell line]</td>\n",
       "      <td>organ site: kidney [cell line]</td>\n",
       "      <td>organ site: kidney [cell line]</td>\n",
       "      <td>organ site: kidney [cell line]</td>\n",
       "      <td>organ site: kidney [cell line]</td>\n",
       "      <td>organ site: prostate [cell line]</td>\n",
       "      <td>organ site: prostate [cell line]</td>\n",
       "      <td>organ site: prostate [cell line]</td>\n",
       "      <td>organ site: prostate [cell line]</td>\n",
       "      <td>organ site: kidney [cell line]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>tissue type: renal cell carcinoma [clear cell ...</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: NA</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: male</td>\n",
       "      <td>gender: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cluster id: B</td>\n",
       "      <td>cluster id: B</td>\n",
       "      <td>cluster id: A</td>\n",
       "      <td>cluster id: C</td>\n",
       "      <td>cluster id: B</td>\n",
       "      <td>cluster id: B</td>\n",
       "      <td>cluster id: A</td>\n",
       "      <td>cluster id: B</td>\n",
       "      <td>cluster id: B</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue type: renal cell carcinoma [cell line S...</td>\n",
       "      <td>tissue type: renal cell carcinoma [cell line C...</td>\n",
       "      <td>tissue type: renal cell carcinoma [cell line S...</td>\n",
       "      <td>tissue type: renal cell carcinoma [cell line K...</td>\n",
       "      <td>tissue type: renal cell carcinoma [cell line S...</td>\n",
       "      <td>tissue type: prostate adenocarcinoma [cell lin...</td>\n",
       "      <td>tissue type: prostate adenocarcinoma [cell lin...</td>\n",
       "      <td>tissue type: prostate adenocarcinoma [cell lin...</td>\n",
       "      <td>tissue type: prostate adenocarcinoma [cell lin...</td>\n",
       "      <td>tissue type: renal cell carcinoma [cell line H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "      <td>cluster id: NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 148 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "5  !Sample_characteristics_ch1   \n",
       "6  !Sample_characteristics_ch1   \n",
       "7  !Sample_characteristics_ch1   \n",
       "8  !Sample_characteristics_ch1   \n",
       "9  !Sample_characteristics_ch1   \n",
       "\n",
       "                                           GSM498450  \\\n",
       "0                                           grade: 2   \n",
       "1                                           stage: 2   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                       gender: male   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: B   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498451  \\\n",
       "0                                           grade: 2   \n",
       "1                                           stage: 2   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                         gender: NA   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: B   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498452  \\\n",
       "0                                           grade: 2   \n",
       "1                                           stage: 2   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                         gender: NA   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: A   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498453  \\\n",
       "0                                           grade: 2   \n",
       "1                                           stage: 1   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                       gender: male   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: C   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498454  \\\n",
       "0                                           grade: 1   \n",
       "1                                           stage: 2   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                         gender: NA   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: B   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498455  \\\n",
       "0                                           grade: 2   \n",
       "1                                           stage: 2   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                         gender: NA   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: B   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498456  \\\n",
       "0                                           grade: 1   \n",
       "1                                           stage: 1   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                         gender: NA   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: A   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498457  \\\n",
       "0                                           grade: 2   \n",
       "1                                           stage: 2   \n",
       "2                             sample type: neoplasia   \n",
       "3                               icd-o 3 code: 8310/3   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...   \n",
       "5                                 organ site: kidney   \n",
       "6                                     gender: female   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...   \n",
       "8                                      cluster id: B   \n",
       "9                                                NaN   \n",
       "\n",
       "                                           GSM498458  ...  \\\n",
       "0                                           grade: 3  ...   \n",
       "1                                           stage: 3  ...   \n",
       "2                             sample type: neoplasia  ...   \n",
       "3                               icd-o 3 code: 8310/3  ...   \n",
       "4  icd-o 3 diagnosis text: clear cell renal cell ...  ...   \n",
       "5                                 organ site: kidney  ...   \n",
       "6                                         gender: NA  ...   \n",
       "7  tissue type: renal cell carcinoma [clear cell ...  ...   \n",
       "8                                      cluster id: B  ...   \n",
       "9                                                NaN  ...   \n",
       "\n",
       "                                           GSM498587  \\\n",
       "0                                   cell line: SLR20   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8312/3   \n",
       "5       icd-o 3 diagnosis text: renal cell carcinoma   \n",
       "6                     organ site: kidney [cell line]   \n",
       "7                                         gender: NA   \n",
       "8  tissue type: renal cell carcinoma [cell line S...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498588  \\\n",
       "0                                  cell line: Caki-2   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8312/3   \n",
       "5       icd-o 3 diagnosis text: renal cell carcinoma   \n",
       "6                     organ site: kidney [cell line]   \n",
       "7                                         gender: NA   \n",
       "8  tissue type: renal cell carcinoma [cell line C...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498589  \\\n",
       "0                                   cell line: SLR21   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8312/3   \n",
       "5       icd-o 3 diagnosis text: renal cell carcinoma   \n",
       "6                     organ site: kidney [cell line]   \n",
       "7                                         gender: NA   \n",
       "8  tissue type: renal cell carcinoma [cell line S...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498590  \\\n",
       "0                                 cell line: KU19-20   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8312/3   \n",
       "5       icd-o 3 diagnosis text: renal cell carcinoma   \n",
       "6                     organ site: kidney [cell line]   \n",
       "7                                         gender: NA   \n",
       "8  tissue type: renal cell carcinoma [cell line K...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498591  \\\n",
       "0                                   cell line: SLR23   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8312/3   \n",
       "5       icd-o 3 diagnosis text: renal cell carcinoma   \n",
       "6                     organ site: kidney [cell line]   \n",
       "7                                         gender: NA   \n",
       "8  tissue type: renal cell carcinoma [cell line S...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498592  \\\n",
       "0                                cell line: PC3hep27   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8140/3   \n",
       "5        icd-o 3 diagnosis text: adenocarcinoma, NOS   \n",
       "6                   organ site: prostate [cell line]   \n",
       "7                                       gender: male   \n",
       "8  tissue type: prostate adenocarcinoma [cell lin...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498593  \\\n",
       "0                                cell line: PC3hep30   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8140/3   \n",
       "5        icd-o 3 diagnosis text: adenocarcinoma, NOS   \n",
       "6                   organ site: prostate [cell line]   \n",
       "7                                       gender: male   \n",
       "8  tissue type: prostate adenocarcinoma [cell lin...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498594  \\\n",
       "0                                 cell line: PC3vec1   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8140/3   \n",
       "5        icd-o 3 diagnosis text: adenocarcinoma, NOS   \n",
       "6                   organ site: prostate [cell line]   \n",
       "7                                       gender: male   \n",
       "8  tissue type: prostate adenocarcinoma [cell lin...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498595  \\\n",
       "0                                 cell line: PC3vec3   \n",
       "1                                          grade: NA   \n",
       "2                                          stage: NA   \n",
       "3                             sample type: neoplasia   \n",
       "4                               icd-o 3 code: 8140/3   \n",
       "5        icd-o 3 diagnosis text: adenocarcinoma, NOS   \n",
       "6                   organ site: prostate [cell line]   \n",
       "7                                       gender: male   \n",
       "8  tissue type: prostate adenocarcinoma [cell lin...   \n",
       "9                                     cluster id: NA   \n",
       "\n",
       "                                           GSM498596  \n",
       "0                                    cell line: HK-2  \n",
       "1                                          grade: NA  \n",
       "2                                          stage: NA  \n",
       "3                             sample type: neoplasia  \n",
       "4                               icd-o 3 code: 8312/3  \n",
       "5       icd-o 3 diagnosis text: renal cell carcinoma  \n",
       "6                     organ site: kidney [cell line]  \n",
       "7                                         gender: NA  \n",
       "8  tissue type: renal cell carcinoma [cell line H...  \n",
       "9                                     cluster id: NA  \n",
       "\n",
       "[10 rows x 148 columns]"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'disease state: --',\n",
       "       'disease state: Leiomyoma', 'disease state: Lung_Adenocarcinoma',\n",
       "       'disease state: Conventional_Clear_Cell_Renal_Cell_Carcinoma',\n",
       "       'disease state: Squamous Cell Carcinoma',\n",
       "       'disease state: Stomach Adenocarcinoma',\n",
       "       'disease state: Large Cell Lymphoma',\n",
       "       'disease state: Malignant Melanoma',\n",
       "       'disease state: Recurrent Renal Cell Carcinoma',\n",
       "       'disease state: Adrenal Cortical Adenoma',\n",
       "       'disease state: Ovarian Adenocarcinoma',\n",
       "       'disease state: Gastrointestinal_Stromal_Tumor',\n",
       "       'disease state: Metastatic Renal Cell Carcinoma',\n",
       "       'disease state: Non neoplastic liver with cirrosis',\n",
       "       'disease state: Malignant G1 Stromal Tumor',\n",
       "       'disease state: melanoma'], dtype=object)"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if (tissue_type == 'disease state: Conventional_Clear_Cell_Renal_Cell_Carcinoma' or tissue_type == 'disease state: Recurrent Renal Cell Carcinoma' or tissue_type == 'disease state: Metastatic Renal Cell Carcinoma'):\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# 这个函数的目的是将年龄的字符串表示转换为一个连续的数值型表示。如果年龄未知（例如，标记为'n.a.'），则返回None。\n",
    "# 函数尝试从传入的字符串中提取出一个整数作为年龄值。如果字符串的格式不符合预期，导致提取失败，同样返回None。\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(':')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# 这个函数将性别的字符串表示转换为二进制值，其中“female”对应1，“male”对应0。如果性别未知或字符串不符合预期格式，则返回None。\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1676864</th>\n",
       "      <th>GSM1676865</th>\n",
       "      <th>GSM1676866</th>\n",
       "      <th>GSM1676867</th>\n",
       "      <th>GSM1676868</th>\n",
       "      <th>GSM1676869</th>\n",
       "      <th>GSM1676870</th>\n",
       "      <th>GSM1676871</th>\n",
       "      <th>GSM1676872</th>\n",
       "      <th>GSM1676873</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1676991</th>\n",
       "      <th>GSM1676992</th>\n",
       "      <th>GSM1676993</th>\n",
       "      <th>GSM1676994</th>\n",
       "      <th>GSM1676995</th>\n",
       "      <th>GSM1676996</th>\n",
       "      <th>GSM1676997</th>\n",
       "      <th>GSM1676998</th>\n",
       "      <th>GSM1676999</th>\n",
       "      <th>GSM1677000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM1676864  GSM1676865  GSM1676866  GSM1676867  \\\n",
       "Kidney Chromophobe           0           0           0           0   \n",
       "\n",
       "                    GSM1676868  GSM1676869  GSM1676870  GSM1676871  \\\n",
       "Kidney Chromophobe           0           0           0           0   \n",
       "\n",
       "                    GSM1676872  GSM1676873  ...  GSM1676991  GSM1676992  \\\n",
       "Kidney Chromophobe           0           0  ...           0           0   \n",
       "\n",
       "                    GSM1676993  GSM1676994  GSM1676995  GSM1676996  \\\n",
       "Kidney Chromophobe           0           0           0           0   \n",
       "\n",
       "                    GSM1676997  GSM1676998  GSM1676999  GSM1677000  \n",
       "Kidney Chromophobe           0           0           0           0  \n",
       "\n",
       "[1 rows x 137 columns]"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1676864</th>\n",
       "      <th>GSM1676865</th>\n",
       "      <th>GSM1676866</th>\n",
       "      <th>GSM1676867</th>\n",
       "      <th>GSM1676868</th>\n",
       "      <th>GSM1676869</th>\n",
       "      <th>GSM1676870</th>\n",
       "      <th>GSM1676871</th>\n",
       "      <th>GSM1676872</th>\n",
       "      <th>GSM1676873</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1676991</th>\n",
       "      <th>GSM1676992</th>\n",
       "      <th>GSM1676993</th>\n",
       "      <th>GSM1676994</th>\n",
       "      <th>GSM1676995</th>\n",
       "      <th>GSM1676996</th>\n",
       "      <th>GSM1676997</th>\n",
       "      <th>GSM1676998</th>\n",
       "      <th>GSM1676999</th>\n",
       "      <th>GSM1677000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>1932.38000</td>\n",
       "      <td>1032.65000</td>\n",
       "      <td>1282.80000</td>\n",
       "      <td>2688.61000</td>\n",
       "      <td>2189.27000</td>\n",
       "      <td>342.63900</td>\n",
       "      <td>254.99600</td>\n",
       "      <td>2225.93000</td>\n",
       "      <td>2785.70000</td>\n",
       "      <td>1785.58000</td>\n",
       "      <td>...</td>\n",
       "      <td>2212.02000</td>\n",
       "      <td>1105.37000</td>\n",
       "      <td>1429.44000</td>\n",
       "      <td>1034.43000</td>\n",
       "      <td>1963.04000</td>\n",
       "      <td>685.84300</td>\n",
       "      <td>1038.43000</td>\n",
       "      <td>267.58600</td>\n",
       "      <td>854.45600</td>\n",
       "      <td>1051.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>833.80500</td>\n",
       "      <td>1034.23000</td>\n",
       "      <td>647.45200</td>\n",
       "      <td>149.05600</td>\n",
       "      <td>315.93400</td>\n",
       "      <td>155.56400</td>\n",
       "      <td>312.14500</td>\n",
       "      <td>431.22300</td>\n",
       "      <td>325.09200</td>\n",
       "      <td>793.21600</td>\n",
       "      <td>...</td>\n",
       "      <td>199.23700</td>\n",
       "      <td>514.35200</td>\n",
       "      <td>879.66300</td>\n",
       "      <td>1314.85000</td>\n",
       "      <td>625.29200</td>\n",
       "      <td>727.30500</td>\n",
       "      <td>521.02900</td>\n",
       "      <td>205.01300</td>\n",
       "      <td>478.05300</td>\n",
       "      <td>466.55800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>122.25500</td>\n",
       "      <td>59.32650</td>\n",
       "      <td>126.75000</td>\n",
       "      <td>139.89900</td>\n",
       "      <td>98.95540</td>\n",
       "      <td>66.88500</td>\n",
       "      <td>74.15530</td>\n",
       "      <td>131.67200</td>\n",
       "      <td>77.37710</td>\n",
       "      <td>120.16700</td>\n",
       "      <td>...</td>\n",
       "      <td>429.30400</td>\n",
       "      <td>101.45700</td>\n",
       "      <td>91.43080</td>\n",
       "      <td>83.34430</td>\n",
       "      <td>197.26200</td>\n",
       "      <td>115.29900</td>\n",
       "      <td>128.30800</td>\n",
       "      <td>84.83030</td>\n",
       "      <td>110.79500</td>\n",
       "      <td>94.88720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>1134.54000</td>\n",
       "      <td>1058.97000</td>\n",
       "      <td>1107.48000</td>\n",
       "      <td>1712.00000</td>\n",
       "      <td>1175.17000</td>\n",
       "      <td>1004.85000</td>\n",
       "      <td>943.42200</td>\n",
       "      <td>1246.45000</td>\n",
       "      <td>1114.25000</td>\n",
       "      <td>1289.73000</td>\n",
       "      <td>...</td>\n",
       "      <td>1237.05000</td>\n",
       "      <td>848.33100</td>\n",
       "      <td>760.27900</td>\n",
       "      <td>608.74700</td>\n",
       "      <td>750.60200</td>\n",
       "      <td>815.26600</td>\n",
       "      <td>822.66700</td>\n",
       "      <td>513.12200</td>\n",
       "      <td>741.42200</td>\n",
       "      <td>630.62300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>26.65850</td>\n",
       "      <td>41.44600</td>\n",
       "      <td>58.22860</td>\n",
       "      <td>77.86010</td>\n",
       "      <td>70.63870</td>\n",
       "      <td>37.91410</td>\n",
       "      <td>16.55080</td>\n",
       "      <td>106.73700</td>\n",
       "      <td>77.79580</td>\n",
       "      <td>67.82810</td>\n",
       "      <td>...</td>\n",
       "      <td>19.03560</td>\n",
       "      <td>16.15000</td>\n",
       "      <td>8.93552</td>\n",
       "      <td>3.17154</td>\n",
       "      <td>93.98570</td>\n",
       "      <td>33.79360</td>\n",
       "      <td>25.44110</td>\n",
       "      <td>25.68730</td>\n",
       "      <td>32.24910</td>\n",
       "      <td>12.31470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <td>7.04814</td>\n",
       "      <td>8.57466</td>\n",
       "      <td>9.11362</td>\n",
       "      <td>12.67890</td>\n",
       "      <td>14.98140</td>\n",
       "      <td>15.68270</td>\n",
       "      <td>5.77798</td>\n",
       "      <td>9.34684</td>\n",
       "      <td>8.36869</td>\n",
       "      <td>8.44274</td>\n",
       "      <td>...</td>\n",
       "      <td>53.56280</td>\n",
       "      <td>7.69069</td>\n",
       "      <td>7.84491</td>\n",
       "      <td>4.28603</td>\n",
       "      <td>24.32590</td>\n",
       "      <td>7.12141</td>\n",
       "      <td>8.55965</td>\n",
       "      <td>18.17220</td>\n",
       "      <td>60.04510</td>\n",
       "      <td>5.14640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <td>27.75430</td>\n",
       "      <td>24.09150</td>\n",
       "      <td>31.00100</td>\n",
       "      <td>25.64010</td>\n",
       "      <td>38.78030</td>\n",
       "      <td>16.97470</td>\n",
       "      <td>25.77740</td>\n",
       "      <td>18.24490</td>\n",
       "      <td>25.75680</td>\n",
       "      <td>28.71170</td>\n",
       "      <td>...</td>\n",
       "      <td>71.87460</td>\n",
       "      <td>26.04780</td>\n",
       "      <td>7.86226</td>\n",
       "      <td>5.50425</td>\n",
       "      <td>65.32950</td>\n",
       "      <td>30.76890</td>\n",
       "      <td>7.79044</td>\n",
       "      <td>33.48380</td>\n",
       "      <td>15.24170</td>\n",
       "      <td>23.51990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <td>3.24531</td>\n",
       "      <td>2.72867</td>\n",
       "      <td>20.23790</td>\n",
       "      <td>19.78960</td>\n",
       "      <td>13.26810</td>\n",
       "      <td>13.99090</td>\n",
       "      <td>3.08212</td>\n",
       "      <td>1.03356</td>\n",
       "      <td>29.03380</td>\n",
       "      <td>3.06343</td>\n",
       "      <td>...</td>\n",
       "      <td>2.39710</td>\n",
       "      <td>16.63280</td>\n",
       "      <td>1.82486</td>\n",
       "      <td>5.90471</td>\n",
       "      <td>5.78538</td>\n",
       "      <td>2.84332</td>\n",
       "      <td>2.08798</td>\n",
       "      <td>16.99070</td>\n",
       "      <td>4.29375</td>\n",
       "      <td>2.71848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <td>7.19261</td>\n",
       "      <td>7.76122</td>\n",
       "      <td>26.20500</td>\n",
       "      <td>12.48630</td>\n",
       "      <td>5.29735</td>\n",
       "      <td>4.88854</td>\n",
       "      <td>28.73010</td>\n",
       "      <td>9.19403</td>\n",
       "      <td>7.81333</td>\n",
       "      <td>43.42730</td>\n",
       "      <td>...</td>\n",
       "      <td>18.31710</td>\n",
       "      <td>27.25600</td>\n",
       "      <td>38.41590</td>\n",
       "      <td>4.80571</td>\n",
       "      <td>60.95720</td>\n",
       "      <td>32.79150</td>\n",
       "      <td>38.91350</td>\n",
       "      <td>7.47526</td>\n",
       "      <td>12.16950</td>\n",
       "      <td>31.59900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <td>8.09025</td>\n",
       "      <td>3.45767</td>\n",
       "      <td>8.66451</td>\n",
       "      <td>7.16433</td>\n",
       "      <td>5.50763</td>\n",
       "      <td>2.13851</td>\n",
       "      <td>7.10520</td>\n",
       "      <td>4.43261</td>\n",
       "      <td>9.89462</td>\n",
       "      <td>6.24205</td>\n",
       "      <td>...</td>\n",
       "      <td>8.93558</td>\n",
       "      <td>5.40206</td>\n",
       "      <td>7.78304</td>\n",
       "      <td>2.96822</td>\n",
       "      <td>6.76194</td>\n",
       "      <td>7.41068</td>\n",
       "      <td>8.06491</td>\n",
       "      <td>9.67596</td>\n",
       "      <td>3.60275</td>\n",
       "      <td>7.40940</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22283 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM1676864  GSM1676865  GSM1676866  GSM1676867  GSM1676868  \\\n",
       "ID                                                                            \n",
       "1007_s_at        1932.38000  1032.65000  1282.80000  2688.61000  2189.27000   \n",
       "1053_at           833.80500  1034.23000   647.45200   149.05600   315.93400   \n",
       "117_at            122.25500    59.32650   126.75000   139.89900    98.95540   \n",
       "121_at           1134.54000  1058.97000  1107.48000  1712.00000  1175.17000   \n",
       "1255_g_at          26.65850    41.44600    58.22860    77.86010    70.63870   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at      7.04814     8.57466     9.11362    12.67890    14.98140   \n",
       "AFFX-ThrX-M_at     27.75430    24.09150    31.00100    25.64010    38.78030   \n",
       "AFFX-TrpnX-3_at     3.24531     2.72867    20.23790    19.78960    13.26810   \n",
       "AFFX-TrpnX-5_at     7.19261     7.76122    26.20500    12.48630     5.29735   \n",
       "AFFX-TrpnX-M_at     8.09025     3.45767     8.66451     7.16433     5.50763   \n",
       "\n",
       "                 GSM1676869  GSM1676870  GSM1676871  GSM1676872  GSM1676873  \\\n",
       "ID                                                                            \n",
       "1007_s_at         342.63900   254.99600  2225.93000  2785.70000  1785.58000   \n",
       "1053_at           155.56400   312.14500   431.22300   325.09200   793.21600   \n",
       "117_at             66.88500    74.15530   131.67200    77.37710   120.16700   \n",
       "121_at           1004.85000   943.42200  1246.45000  1114.25000  1289.73000   \n",
       "1255_g_at          37.91410    16.55080   106.73700    77.79580    67.82810   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     15.68270     5.77798     9.34684     8.36869     8.44274   \n",
       "AFFX-ThrX-M_at     16.97470    25.77740    18.24490    25.75680    28.71170   \n",
       "AFFX-TrpnX-3_at    13.99090     3.08212     1.03356    29.03380     3.06343   \n",
       "AFFX-TrpnX-5_at     4.88854    28.73010     9.19403     7.81333    43.42730   \n",
       "AFFX-TrpnX-M_at     2.13851     7.10520     4.43261     9.89462     6.24205   \n",
       "\n",
       "                 ...  GSM1676991  GSM1676992  GSM1676993  GSM1676994  \\\n",
       "ID               ...                                                   \n",
       "1007_s_at        ...  2212.02000  1105.37000  1429.44000  1034.43000   \n",
       "1053_at          ...   199.23700   514.35200   879.66300  1314.85000   \n",
       "117_at           ...   429.30400   101.45700    91.43080    83.34430   \n",
       "121_at           ...  1237.05000   848.33100   760.27900   608.74700   \n",
       "1255_g_at        ...    19.03560    16.15000     8.93552     3.17154   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at   ...    53.56280     7.69069     7.84491     4.28603   \n",
       "AFFX-ThrX-M_at   ...    71.87460    26.04780     7.86226     5.50425   \n",
       "AFFX-TrpnX-3_at  ...     2.39710    16.63280     1.82486     5.90471   \n",
       "AFFX-TrpnX-5_at  ...    18.31710    27.25600    38.41590     4.80571   \n",
       "AFFX-TrpnX-M_at  ...     8.93558     5.40206     7.78304     2.96822   \n",
       "\n",
       "                 GSM1676995  GSM1676996  GSM1676997  GSM1676998  GSM1676999  \\\n",
       "ID                                                                            \n",
       "1007_s_at        1963.04000   685.84300  1038.43000   267.58600   854.45600   \n",
       "1053_at           625.29200   727.30500   521.02900   205.01300   478.05300   \n",
       "117_at            197.26200   115.29900   128.30800    84.83030   110.79500   \n",
       "121_at            750.60200   815.26600   822.66700   513.12200   741.42200   \n",
       "1255_g_at          93.98570    33.79360    25.44110    25.68730    32.24910   \n",
       "...                     ...         ...         ...         ...         ...   \n",
       "AFFX-ThrX-5_at     24.32590     7.12141     8.55965    18.17220    60.04510   \n",
       "AFFX-ThrX-M_at     65.32950    30.76890     7.79044    33.48380    15.24170   \n",
       "AFFX-TrpnX-3_at     5.78538     2.84332     2.08798    16.99070     4.29375   \n",
       "AFFX-TrpnX-5_at    60.95720    32.79150    38.91350     7.47526    12.16950   \n",
       "AFFX-TrpnX-M_at     6.76194     7.41068     8.06491     9.67596     3.60275   \n",
       "\n",
       "                 GSM1677000  \n",
       "ID                           \n",
       "1007_s_at        1051.03000  \n",
       "1053_at           466.55800  \n",
       "117_at             94.88720  \n",
       "121_at            630.62300  \n",
       "1255_g_at          12.31470  \n",
       "...                     ...  \n",
       "AFFX-ThrX-5_at      5.14640  \n",
       "AFFX-ThrX-M_at     23.51990  \n",
       "AFFX-TrpnX-3_at     2.71848  \n",
       "AFFX-TrpnX-5_at    31.59900  \n",
       "AFFX-TrpnX-M_at     7.40940  \n",
       "\n",
       "[22283 rows x 137 columns]"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1598_g_at', '160020_at', '1729_at', '1773_at', '177_at', '179_at', '1861_at']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1676864</th>\n",
       "      <th>GSM1676865</th>\n",
       "      <th>GSM1676866</th>\n",
       "      <th>GSM1676867</th>\n",
       "      <th>GSM1676868</th>\n",
       "      <th>GSM1676869</th>\n",
       "      <th>GSM1676870</th>\n",
       "      <th>GSM1676871</th>\n",
       "      <th>GSM1676872</th>\n",
       "      <th>GSM1676873</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1676991</th>\n",
       "      <th>GSM1676992</th>\n",
       "      <th>GSM1676993</th>\n",
       "      <th>GSM1676994</th>\n",
       "      <th>GSM1676995</th>\n",
       "      <th>GSM1676996</th>\n",
       "      <th>GSM1676997</th>\n",
       "      <th>GSM1676998</th>\n",
       "      <th>GSM1676999</th>\n",
       "      <th>GSM1677000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>411.658000</td>\n",
       "      <td>328.766000</td>\n",
       "      <td>283.038000</td>\n",
       "      <td>441.273000</td>\n",
       "      <td>456.17000</td>\n",
       "      <td>356.891000</td>\n",
       "      <td>370.538000</td>\n",
       "      <td>391.200000</td>\n",
       "      <td>517.63200</td>\n",
       "      <td>330.712000</td>\n",
       "      <td>...</td>\n",
       "      <td>492.9340</td>\n",
       "      <td>252.68000</td>\n",
       "      <td>214.051000</td>\n",
       "      <td>172.401000</td>\n",
       "      <td>360.6670</td>\n",
       "      <td>224.95800</td>\n",
       "      <td>196.6330</td>\n",
       "      <td>423.316000</td>\n",
       "      <td>604.971000</td>\n",
       "      <td>663.523000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.933690</td>\n",
       "      <td>29.729700</td>\n",
       "      <td>31.517800</td>\n",
       "      <td>117.207000</td>\n",
       "      <td>66.05450</td>\n",
       "      <td>76.368500</td>\n",
       "      <td>8.343650</td>\n",
       "      <td>77.880200</td>\n",
       "      <td>96.81630</td>\n",
       "      <td>76.772900</td>\n",
       "      <td>...</td>\n",
       "      <td>1938.5400</td>\n",
       "      <td>34.16810</td>\n",
       "      <td>41.075100</td>\n",
       "      <td>78.744600</td>\n",
       "      <td>13.4254</td>\n",
       "      <td>113.81800</td>\n",
       "      <td>80.4261</td>\n",
       "      <td>90.634000</td>\n",
       "      <td>1469.130000</td>\n",
       "      <td>1450.950000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>67.969200</td>\n",
       "      <td>42.261300</td>\n",
       "      <td>103.195000</td>\n",
       "      <td>34.880400</td>\n",
       "      <td>37.44720</td>\n",
       "      <td>20.832300</td>\n",
       "      <td>63.035600</td>\n",
       "      <td>70.497400</td>\n",
       "      <td>25.55710</td>\n",
       "      <td>77.666600</td>\n",
       "      <td>...</td>\n",
       "      <td>73.6854</td>\n",
       "      <td>85.04110</td>\n",
       "      <td>131.085000</td>\n",
       "      <td>155.877000</td>\n",
       "      <td>138.5640</td>\n",
       "      <td>170.43300</td>\n",
       "      <td>96.1404</td>\n",
       "      <td>115.264000</td>\n",
       "      <td>131.866000</td>\n",
       "      <td>22.694400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>103.083000</td>\n",
       "      <td>19.840200</td>\n",
       "      <td>120.743000</td>\n",
       "      <td>233.940000</td>\n",
       "      <td>132.97200</td>\n",
       "      <td>54.878500</td>\n",
       "      <td>86.099100</td>\n",
       "      <td>86.712800</td>\n",
       "      <td>132.28300</td>\n",
       "      <td>123.381000</td>\n",
       "      <td>...</td>\n",
       "      <td>375.3930</td>\n",
       "      <td>129.99600</td>\n",
       "      <td>115.976000</td>\n",
       "      <td>82.167800</td>\n",
       "      <td>157.9120</td>\n",
       "      <td>148.86000</td>\n",
       "      <td>60.4659</td>\n",
       "      <td>58.103800</td>\n",
       "      <td>47.507900</td>\n",
       "      <td>122.016000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>253.048000</td>\n",
       "      <td>307.313000</td>\n",
       "      <td>301.599000</td>\n",
       "      <td>54.338500</td>\n",
       "      <td>292.66000</td>\n",
       "      <td>526.452000</td>\n",
       "      <td>285.639000</td>\n",
       "      <td>242.461000</td>\n",
       "      <td>157.34800</td>\n",
       "      <td>88.740300</td>\n",
       "      <td>...</td>\n",
       "      <td>168.3570</td>\n",
       "      <td>331.88900</td>\n",
       "      <td>455.359000</td>\n",
       "      <td>512.032000</td>\n",
       "      <td>378.5850</td>\n",
       "      <td>584.66500</td>\n",
       "      <td>611.1190</td>\n",
       "      <td>200.164000</td>\n",
       "      <td>299.390000</td>\n",
       "      <td>332.704000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDB</th>\n",
       "      <td>7.021830</td>\n",
       "      <td>25.621200</td>\n",
       "      <td>49.129700</td>\n",
       "      <td>56.345100</td>\n",
       "      <td>52.64970</td>\n",
       "      <td>9.202230</td>\n",
       "      <td>57.596900</td>\n",
       "      <td>38.391100</td>\n",
       "      <td>56.36770</td>\n",
       "      <td>7.179120</td>\n",
       "      <td>...</td>\n",
       "      <td>12.7427</td>\n",
       "      <td>82.66240</td>\n",
       "      <td>84.754400</td>\n",
       "      <td>15.316100</td>\n",
       "      <td>56.1309</td>\n",
       "      <td>43.45840</td>\n",
       "      <td>32.1018</td>\n",
       "      <td>13.704900</td>\n",
       "      <td>66.226700</td>\n",
       "      <td>72.470700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>245.770000</td>\n",
       "      <td>223.404000</td>\n",
       "      <td>324.696000</td>\n",
       "      <td>408.354000</td>\n",
       "      <td>274.03900</td>\n",
       "      <td>207.566000</td>\n",
       "      <td>41.970100</td>\n",
       "      <td>96.083800</td>\n",
       "      <td>204.60200</td>\n",
       "      <td>277.355000</td>\n",
       "      <td>...</td>\n",
       "      <td>293.5240</td>\n",
       "      <td>105.07400</td>\n",
       "      <td>239.035000</td>\n",
       "      <td>242.894000</td>\n",
       "      <td>112.4240</td>\n",
       "      <td>226.59100</td>\n",
       "      <td>263.2760</td>\n",
       "      <td>15.799600</td>\n",
       "      <td>209.634000</td>\n",
       "      <td>202.095000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>864.049500</td>\n",
       "      <td>361.164500</td>\n",
       "      <td>368.090000</td>\n",
       "      <td>509.184500</td>\n",
       "      <td>699.28350</td>\n",
       "      <td>711.235500</td>\n",
       "      <td>831.348000</td>\n",
       "      <td>898.541000</td>\n",
       "      <td>529.82600</td>\n",
       "      <td>535.816500</td>\n",
       "      <td>...</td>\n",
       "      <td>956.2580</td>\n",
       "      <td>1056.60850</td>\n",
       "      <td>1895.635000</td>\n",
       "      <td>513.773000</td>\n",
       "      <td>1485.5750</td>\n",
       "      <td>625.40500</td>\n",
       "      <td>2031.3400</td>\n",
       "      <td>1590.970000</td>\n",
       "      <td>872.288000</td>\n",
       "      <td>982.831000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>95.892933</td>\n",
       "      <td>186.514933</td>\n",
       "      <td>183.754067</td>\n",
       "      <td>259.034713</td>\n",
       "      <td>168.38015</td>\n",
       "      <td>311.179743</td>\n",
       "      <td>224.061743</td>\n",
       "      <td>125.673357</td>\n",
       "      <td>188.30636</td>\n",
       "      <td>118.847337</td>\n",
       "      <td>...</td>\n",
       "      <td>187.8084</td>\n",
       "      <td>105.01919</td>\n",
       "      <td>100.009013</td>\n",
       "      <td>136.908567</td>\n",
       "      <td>82.2776</td>\n",
       "      <td>180.19429</td>\n",
       "      <td>119.2507</td>\n",
       "      <td>192.480833</td>\n",
       "      <td>127.771333</td>\n",
       "      <td>95.924913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>972.634000</td>\n",
       "      <td>1135.550000</td>\n",
       "      <td>818.291000</td>\n",
       "      <td>659.049000</td>\n",
       "      <td>615.55700</td>\n",
       "      <td>1142.090000</td>\n",
       "      <td>1197.570000</td>\n",
       "      <td>1057.490000</td>\n",
       "      <td>859.80700</td>\n",
       "      <td>929.594000</td>\n",
       "      <td>...</td>\n",
       "      <td>348.7130</td>\n",
       "      <td>389.84600</td>\n",
       "      <td>656.539000</td>\n",
       "      <td>750.973000</td>\n",
       "      <td>778.6430</td>\n",
       "      <td>707.98900</td>\n",
       "      <td>713.4530</td>\n",
       "      <td>765.151000</td>\n",
       "      <td>486.239000</td>\n",
       "      <td>487.393000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11727 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM1676864   GSM1676865  GSM1676866  GSM1676867  GSM1676868  \\\n",
       "A1CF    411.658000   328.766000  283.038000  441.273000   456.17000   \n",
       "A2M       9.933690    29.729700   31.517800  117.207000    66.05450   \n",
       "A4GALT   67.969200    42.261300  103.195000   34.880400    37.44720   \n",
       "A4GNT   103.083000    19.840200  120.743000  233.940000   132.97200   \n",
       "AAAS    253.048000   307.313000  301.599000   54.338500   292.66000   \n",
       "...            ...          ...         ...         ...         ...   \n",
       "ZXDB      7.021830    25.621200   49.129700   56.345100    52.64970   \n",
       "ZXDC    245.770000   223.404000  324.696000  408.354000   274.03900   \n",
       "ZYX     864.049500   361.164500  368.090000  509.184500   699.28350   \n",
       "ZZEF1    95.892933   186.514933  183.754067  259.034713   168.38015   \n",
       "ZZZ3    972.634000  1135.550000  818.291000  659.049000   615.55700   \n",
       "\n",
       "         GSM1676869   GSM1676870   GSM1676871  GSM1676872  GSM1676873  ...  \\\n",
       "A1CF     356.891000   370.538000   391.200000   517.63200  330.712000  ...   \n",
       "A2M       76.368500     8.343650    77.880200    96.81630   76.772900  ...   \n",
       "A4GALT    20.832300    63.035600    70.497400    25.55710   77.666600  ...   \n",
       "A4GNT     54.878500    86.099100    86.712800   132.28300  123.381000  ...   \n",
       "AAAS     526.452000   285.639000   242.461000   157.34800   88.740300  ...   \n",
       "...             ...          ...          ...         ...         ...  ...   \n",
       "ZXDB       9.202230    57.596900    38.391100    56.36770    7.179120  ...   \n",
       "ZXDC     207.566000    41.970100    96.083800   204.60200  277.355000  ...   \n",
       "ZYX      711.235500   831.348000   898.541000   529.82600  535.816500  ...   \n",
       "ZZEF1    311.179743   224.061743   125.673357   188.30636  118.847337  ...   \n",
       "ZZZ3    1142.090000  1197.570000  1057.490000   859.80700  929.594000  ...   \n",
       "\n",
       "        GSM1676991  GSM1676992   GSM1676993  GSM1676994  GSM1676995  \\\n",
       "A1CF      492.9340   252.68000   214.051000  172.401000    360.6670   \n",
       "A2M      1938.5400    34.16810    41.075100   78.744600     13.4254   \n",
       "A4GALT     73.6854    85.04110   131.085000  155.877000    138.5640   \n",
       "A4GNT     375.3930   129.99600   115.976000   82.167800    157.9120   \n",
       "AAAS      168.3570   331.88900   455.359000  512.032000    378.5850   \n",
       "...            ...         ...          ...         ...         ...   \n",
       "ZXDB       12.7427    82.66240    84.754400   15.316100     56.1309   \n",
       "ZXDC      293.5240   105.07400   239.035000  242.894000    112.4240   \n",
       "ZYX       956.2580  1056.60850  1895.635000  513.773000   1485.5750   \n",
       "ZZEF1     187.8084   105.01919   100.009013  136.908567     82.2776   \n",
       "ZZZ3      348.7130   389.84600   656.539000  750.973000    778.6430   \n",
       "\n",
       "        GSM1676996  GSM1676997   GSM1676998   GSM1676999   GSM1677000  \n",
       "A1CF     224.95800    196.6330   423.316000   604.971000   663.523000  \n",
       "A2M      113.81800     80.4261    90.634000  1469.130000  1450.950000  \n",
       "A4GALT   170.43300     96.1404   115.264000   131.866000    22.694400  \n",
       "A4GNT    148.86000     60.4659    58.103800    47.507900   122.016000  \n",
       "AAAS     584.66500    611.1190   200.164000   299.390000   332.704000  \n",
       "...            ...         ...          ...          ...          ...  \n",
       "ZXDB      43.45840     32.1018    13.704900    66.226700    72.470700  \n",
       "ZXDC     226.59100    263.2760    15.799600   209.634000   202.095000  \n",
       "ZYX      625.40500   2031.3400  1590.970000   872.288000   982.831000  \n",
       "ZZEF1    180.19429    119.2507   192.480833   127.771333    95.924913  \n",
       "ZZZ3     707.98900    713.4530   765.151000   486.239000   487.393000  \n",
       "\n",
       "[11727 rows x 137 columns]"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZSWIM8</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1676864</th>\n",
       "      <td>0.0</td>\n",
       "      <td>411.658</td>\n",
       "      <td>9.93369</td>\n",
       "      <td>67.9692</td>\n",
       "      <td>103.0830</td>\n",
       "      <td>253.0480</td>\n",
       "      <td>693.969</td>\n",
       "      <td>52.00740</td>\n",
       "      <td>254.15790</td>\n",
       "      <td>169.690144</td>\n",
       "      <td>...</td>\n",
       "      <td>172.9030</td>\n",
       "      <td>118.234115</td>\n",
       "      <td>1172.070</td>\n",
       "      <td>937.975</td>\n",
       "      <td>4273.62</td>\n",
       "      <td>7.02183</td>\n",
       "      <td>245.7700</td>\n",
       "      <td>864.0495</td>\n",
       "      <td>95.892933</td>\n",
       "      <td>972.634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676865</th>\n",
       "      <td>0.0</td>\n",
       "      <td>328.766</td>\n",
       "      <td>29.72970</td>\n",
       "      <td>42.2613</td>\n",
       "      <td>19.8402</td>\n",
       "      <td>307.3130</td>\n",
       "      <td>565.707</td>\n",
       "      <td>52.20780</td>\n",
       "      <td>411.08050</td>\n",
       "      <td>138.574204</td>\n",
       "      <td>...</td>\n",
       "      <td>160.9360</td>\n",
       "      <td>164.939945</td>\n",
       "      <td>888.014</td>\n",
       "      <td>810.915</td>\n",
       "      <td>6408.85</td>\n",
       "      <td>25.62120</td>\n",
       "      <td>223.4040</td>\n",
       "      <td>361.1645</td>\n",
       "      <td>186.514933</td>\n",
       "      <td>1135.550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676866</th>\n",
       "      <td>0.0</td>\n",
       "      <td>283.038</td>\n",
       "      <td>31.51780</td>\n",
       "      <td>103.1950</td>\n",
       "      <td>120.7430</td>\n",
       "      <td>301.5990</td>\n",
       "      <td>614.077</td>\n",
       "      <td>20.14540</td>\n",
       "      <td>301.78795</td>\n",
       "      <td>171.979780</td>\n",
       "      <td>...</td>\n",
       "      <td>149.0170</td>\n",
       "      <td>182.287700</td>\n",
       "      <td>970.310</td>\n",
       "      <td>769.493</td>\n",
       "      <td>7430.47</td>\n",
       "      <td>49.12970</td>\n",
       "      <td>324.6960</td>\n",
       "      <td>368.0900</td>\n",
       "      <td>183.754067</td>\n",
       "      <td>818.291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676867</th>\n",
       "      <td>0.0</td>\n",
       "      <td>441.273</td>\n",
       "      <td>117.20700</td>\n",
       "      <td>34.8804</td>\n",
       "      <td>233.9400</td>\n",
       "      <td>54.3385</td>\n",
       "      <td>1193.600</td>\n",
       "      <td>26.59360</td>\n",
       "      <td>195.02295</td>\n",
       "      <td>411.052640</td>\n",
       "      <td>...</td>\n",
       "      <td>206.7570</td>\n",
       "      <td>454.626050</td>\n",
       "      <td>910.911</td>\n",
       "      <td>274.856</td>\n",
       "      <td>5882.10</td>\n",
       "      <td>56.34510</td>\n",
       "      <td>408.3540</td>\n",
       "      <td>509.1845</td>\n",
       "      <td>259.034713</td>\n",
       "      <td>659.049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676868</th>\n",
       "      <td>0.0</td>\n",
       "      <td>456.170</td>\n",
       "      <td>66.05450</td>\n",
       "      <td>37.4472</td>\n",
       "      <td>132.9720</td>\n",
       "      <td>292.6600</td>\n",
       "      <td>869.968</td>\n",
       "      <td>77.00920</td>\n",
       "      <td>291.42050</td>\n",
       "      <td>225.582720</td>\n",
       "      <td>...</td>\n",
       "      <td>188.0690</td>\n",
       "      <td>336.587970</td>\n",
       "      <td>796.490</td>\n",
       "      <td>342.181</td>\n",
       "      <td>4632.58</td>\n",
       "      <td>52.64970</td>\n",
       "      <td>274.0390</td>\n",
       "      <td>699.2835</td>\n",
       "      <td>168.380150</td>\n",
       "      <td>615.557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676996</th>\n",
       "      <td>0.0</td>\n",
       "      <td>224.958</td>\n",
       "      <td>113.81800</td>\n",
       "      <td>170.4330</td>\n",
       "      <td>148.8600</td>\n",
       "      <td>584.6650</td>\n",
       "      <td>523.556</td>\n",
       "      <td>7.43009</td>\n",
       "      <td>450.45955</td>\n",
       "      <td>134.190732</td>\n",
       "      <td>...</td>\n",
       "      <td>188.5000</td>\n",
       "      <td>199.339950</td>\n",
       "      <td>657.216</td>\n",
       "      <td>1048.500</td>\n",
       "      <td>5762.04</td>\n",
       "      <td>43.45840</td>\n",
       "      <td>226.5910</td>\n",
       "      <td>625.4050</td>\n",
       "      <td>180.194290</td>\n",
       "      <td>707.989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676997</th>\n",
       "      <td>0.0</td>\n",
       "      <td>196.633</td>\n",
       "      <td>80.42610</td>\n",
       "      <td>96.1404</td>\n",
       "      <td>60.4659</td>\n",
       "      <td>611.1190</td>\n",
       "      <td>447.764</td>\n",
       "      <td>11.25050</td>\n",
       "      <td>342.95000</td>\n",
       "      <td>108.845170</td>\n",
       "      <td>...</td>\n",
       "      <td>205.8840</td>\n",
       "      <td>121.243740</td>\n",
       "      <td>685.352</td>\n",
       "      <td>1352.300</td>\n",
       "      <td>2760.27</td>\n",
       "      <td>32.10180</td>\n",
       "      <td>263.2760</td>\n",
       "      <td>2031.3400</td>\n",
       "      <td>119.250700</td>\n",
       "      <td>713.453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676998</th>\n",
       "      <td>0.0</td>\n",
       "      <td>423.316</td>\n",
       "      <td>90.63400</td>\n",
       "      <td>115.2640</td>\n",
       "      <td>58.1038</td>\n",
       "      <td>200.1640</td>\n",
       "      <td>860.323</td>\n",
       "      <td>9.11952</td>\n",
       "      <td>255.18895</td>\n",
       "      <td>120.683550</td>\n",
       "      <td>...</td>\n",
       "      <td>89.2204</td>\n",
       "      <td>201.062600</td>\n",
       "      <td>441.399</td>\n",
       "      <td>575.010</td>\n",
       "      <td>1340.71</td>\n",
       "      <td>13.70490</td>\n",
       "      <td>15.7996</td>\n",
       "      <td>1590.9700</td>\n",
       "      <td>192.480833</td>\n",
       "      <td>765.151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1676999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>604.971</td>\n",
       "      <td>1469.13000</td>\n",
       "      <td>131.8660</td>\n",
       "      <td>47.5079</td>\n",
       "      <td>299.3900</td>\n",
       "      <td>596.813</td>\n",
       "      <td>63.87370</td>\n",
       "      <td>294.99555</td>\n",
       "      <td>163.793346</td>\n",
       "      <td>...</td>\n",
       "      <td>135.0100</td>\n",
       "      <td>170.178150</td>\n",
       "      <td>469.864</td>\n",
       "      <td>1394.440</td>\n",
       "      <td>3616.59</td>\n",
       "      <td>66.22670</td>\n",
       "      <td>209.6340</td>\n",
       "      <td>872.2880</td>\n",
       "      <td>127.771333</td>\n",
       "      <td>486.239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1677000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>663.523</td>\n",
       "      <td>1450.95000</td>\n",
       "      <td>22.6944</td>\n",
       "      <td>122.0160</td>\n",
       "      <td>332.7040</td>\n",
       "      <td>514.074</td>\n",
       "      <td>9.41142</td>\n",
       "      <td>296.73410</td>\n",
       "      <td>145.633336</td>\n",
       "      <td>...</td>\n",
       "      <td>131.1670</td>\n",
       "      <td>183.589250</td>\n",
       "      <td>492.173</td>\n",
       "      <td>1160.180</td>\n",
       "      <td>3245.00</td>\n",
       "      <td>72.47070</td>\n",
       "      <td>202.0950</td>\n",
       "      <td>982.8310</td>\n",
       "      <td>95.924913</td>\n",
       "      <td>487.393</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows × 11728 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Kidney Chromophobe     A1CF         A2M    A4GALT     A4GNT  \\\n",
       "GSM1676864                 0.0  411.658     9.93369   67.9692  103.0830   \n",
       "GSM1676865                 0.0  328.766    29.72970   42.2613   19.8402   \n",
       "GSM1676866                 0.0  283.038    31.51780  103.1950  120.7430   \n",
       "GSM1676867                 0.0  441.273   117.20700   34.8804  233.9400   \n",
       "GSM1676868                 0.0  456.170    66.05450   37.4472  132.9720   \n",
       "...                        ...      ...         ...       ...       ...   \n",
       "GSM1676996                 0.0  224.958   113.81800  170.4330  148.8600   \n",
       "GSM1676997                 0.0  196.633    80.42610   96.1404   60.4659   \n",
       "GSM1676998                 0.0  423.316    90.63400  115.2640   58.1038   \n",
       "GSM1676999                 0.0  604.971  1469.13000  131.8660   47.5079   \n",
       "GSM1677000                 0.0  663.523  1450.95000   22.6944  122.0160   \n",
       "\n",
       "                AAAS      AACS     AADAC      AAGAB        AAK1  ...  \\\n",
       "GSM1676864  253.0480   693.969  52.00740  254.15790  169.690144  ...   \n",
       "GSM1676865  307.3130   565.707  52.20780  411.08050  138.574204  ...   \n",
       "GSM1676866  301.5990   614.077  20.14540  301.78795  171.979780  ...   \n",
       "GSM1676867   54.3385  1193.600  26.59360  195.02295  411.052640  ...   \n",
       "GSM1676868  292.6600   869.968  77.00920  291.42050  225.582720  ...   \n",
       "...              ...       ...       ...        ...         ...  ...   \n",
       "GSM1676996  584.6650   523.556   7.43009  450.45955  134.190732  ...   \n",
       "GSM1676997  611.1190   447.764  11.25050  342.95000  108.845170  ...   \n",
       "GSM1676998  200.1640   860.323   9.11952  255.18895  120.683550  ...   \n",
       "GSM1676999  299.3900   596.813  63.87370  294.99555  163.793346  ...   \n",
       "GSM1677000  332.7040   514.074   9.41142  296.73410  145.633336  ...   \n",
       "\n",
       "              ZSWIM1      ZSWIM8      ZW10    ZWILCH    ZWINT      ZXDB  \\\n",
       "GSM1676864  172.9030  118.234115  1172.070   937.975  4273.62   7.02183   \n",
       "GSM1676865  160.9360  164.939945   888.014   810.915  6408.85  25.62120   \n",
       "GSM1676866  149.0170  182.287700   970.310   769.493  7430.47  49.12970   \n",
       "GSM1676867  206.7570  454.626050   910.911   274.856  5882.10  56.34510   \n",
       "GSM1676868  188.0690  336.587970   796.490   342.181  4632.58  52.64970   \n",
       "...              ...         ...       ...       ...      ...       ...   \n",
       "GSM1676996  188.5000  199.339950   657.216  1048.500  5762.04  43.45840   \n",
       "GSM1676997  205.8840  121.243740   685.352  1352.300  2760.27  32.10180   \n",
       "GSM1676998   89.2204  201.062600   441.399   575.010  1340.71  13.70490   \n",
       "GSM1676999  135.0100  170.178150   469.864  1394.440  3616.59  66.22670   \n",
       "GSM1677000  131.1670  183.589250   492.173  1160.180  3245.00  72.47070   \n",
       "\n",
       "                ZXDC        ZYX       ZZEF1      ZZZ3  \n",
       "GSM1676864  245.7700   864.0495   95.892933   972.634  \n",
       "GSM1676865  223.4040   361.1645  186.514933  1135.550  \n",
       "GSM1676866  324.6960   368.0900  183.754067   818.291  \n",
       "GSM1676867  408.3540   509.1845  259.034713   659.049  \n",
       "GSM1676868  274.0390   699.2835  168.380150   615.557  \n",
       "...              ...        ...         ...       ...  \n",
       "GSM1676996  226.5910   625.4050  180.194290   707.989  \n",
       "GSM1676997  263.2760  2031.3400  119.250700   713.453  \n",
       "GSM1676998   15.7996  1590.9700  192.480833   765.151  \n",
       "GSM1676999  209.6340   872.2880  127.771333   486.239  \n",
       "GSM1677000  202.0950   982.8310   95.924913   487.393  \n",
       "\n",
       "[137 rows x 11728 columns]"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 137 samples.\n",
      "For the feature 'Kidney Chromophobe', the least common label is '1.0' with 11 occurrences. This represents 8.03% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"caArray_dobbi-00100: Interlaboratory comparability study of cancer gene expression analysis using oligonucleotide microarrays\"\n",
      "!Series_summary\t\"A key step in bringing gene expression data into clinical practice is the conduct of large studies to confirm preliminary models. The performance of such confirmatory studies and the transition to clinical practice requires that microarray data from different laboratories are comparable and reproducible. We designed a study to assess the comparability of data from four laboratories that will conduct a larger microarray profiling confirmation project in lung adenocarcinomas. To test the feasibility of combining data across laboratories, frozen tumor tissues, cell line pellets, and purified RNA samples were analyzed at each of the four laboratories. Samples of each type and several subsamples from each tumor and each cell line were blinded before being distributed. The laboratories followed a common protocol for all steps of tissue processing, RNA extraction, and microarray analysis using Affymetrix Human Genome U133A arrays. High within-laboratory and between-laboratory correlations were observed on the purified RNA samples, the cell lines, and the frozen tumor tissues. Intraclass correlation within laboratories was only slightly stronger than between laboratories, and the intraclass correlation tended to be weakest for genes expressed at low levels and showing small variation. Finally, hierarchical cluster analysis revealed that the repeated samples clustered together regardless of the laboratory in which the experiments were done. The findings indicate that under properly controlled conditions it is feasible to perform complete tumor microarray analysis, from tissue processing to hybridization and scanning, at multiple independent laboratories for a single study.\"\n",
      "!Series_overall_design\t\"dobbi-00100\"\n",
      "!Series_overall_design\t\"Assay Type: Gene Expression\"\n",
      "!Series_overall_design\t\"Provider: Affymetrix\"\n",
      "!Series_overall_design\t\"Array Designs: HG-U133A\"\n",
      "!Series_overall_design\t\"Organism: Homo sapiens (ncbitax)\"\n",
      "!Series_overall_design\t\"Tissue Sites: Kidney, Lung, Stomach, Uterus, Liver, Lymphoid tissue, Ovary, Skin, Adrenal Gland, Lymph_Node\"\n",
      "!Series_overall_design\t\"Material Types: cell, nuclear_RNA, synthetic_RNA, organism_part, total_RNA\"\n",
      "!Series_overall_design\t\"Disease States: Recurrent Renal Cell Carcinoma, Squamous Cell Carcinoma,Conventional_Clear_Cell_Renal_Cell_Carcinoma,Gastrointestinal_Stromal_Tumor, Lung_Adenocarcinoma, Leiomyoma, Non neoplastic liver with cirrosis, Stomach Adenocarcinoma, Large Cell Lymphoma, Ovarian Adenocarcinoma, Melanoma, Malignant G1 Stromal Tumor, Adrenal Cortical Adenoma, Metastatic Renal Cell Carcinoma, Malignant Melanoma\"\n"
     ]
    }
   ],
   "source": [
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1676864</th>\n",
       "      <th>GSM1676865</th>\n",
       "      <th>GSM1676866</th>\n",
       "      <th>GSM1676867</th>\n",
       "      <th>GSM1676868</th>\n",
       "      <th>GSM1676869</th>\n",
       "      <th>GSM1676870</th>\n",
       "      <th>GSM1676871</th>\n",
       "      <th>GSM1676872</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1676991</th>\n",
       "      <th>GSM1676992</th>\n",
       "      <th>GSM1676993</th>\n",
       "      <th>GSM1676994</th>\n",
       "      <th>GSM1676995</th>\n",
       "      <th>GSM1676996</th>\n",
       "      <th>GSM1676997</th>\n",
       "      <th>GSM1676998</th>\n",
       "      <th>GSM1676999</th>\n",
       "      <th>GSM1677000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell line: H2347</td>\n",
       "      <td>cell line: H2347</td>\n",
       "      <td>cell line: H1437</td>\n",
       "      <td>cell line: HCC78</td>\n",
       "      <td>cell line: HCC78</td>\n",
       "      <td>cell line: H2087</td>\n",
       "      <td>cell line: H2087</td>\n",
       "      <td>cell line: H2009</td>\n",
       "      <td>cell line: H2009</td>\n",
       "      <td>...</td>\n",
       "      <td>cell line: --</td>\n",
       "      <td>cell line: HCC78</td>\n",
       "      <td>cell line: H2347</td>\n",
       "      <td>cell line: H1437</td>\n",
       "      <td>cell line: H2009</td>\n",
       "      <td>cell line: H1437</td>\n",
       "      <td>cell line: H2347</td>\n",
       "      <td>cell line: H2087</td>\n",
       "      <td>cell line: --</td>\n",
       "      <td>cell line: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: Leiomyoma</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: Squamous Cell Carcinoma</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "      <td>disease state: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>...</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "      <td>tumor grading: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>...</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "      <td>disease stage: --</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: Uterus</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>...</td>\n",
       "      <td>organism part: Lung</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "      <td>organism part: --</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 138 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession         GSM1676864         GSM1676865  \\\n",
       "0  !Sample_characteristics_ch1   cell line: H2347   cell line: H2347   \n",
       "1  !Sample_characteristics_ch1  disease state: --  disease state: --   \n",
       "2  !Sample_characteristics_ch1  tumor grading: --  tumor grading: --   \n",
       "3  !Sample_characteristics_ch1  disease stage: --  disease stage: --   \n",
       "4  !Sample_characteristics_ch1  organism part: --  organism part: --   \n",
       "\n",
       "                 GSM1676866         GSM1676867         GSM1676868  \\\n",
       "0          cell line: H1437   cell line: HCC78   cell line: HCC78   \n",
       "1  disease state: Leiomyoma  disease state: --  disease state: --   \n",
       "2         tumor grading: --  tumor grading: --  tumor grading: --   \n",
       "3         disease stage: --  disease stage: --  disease stage: --   \n",
       "4     organism part: Uterus  organism part: --  organism part: --   \n",
       "\n",
       "          GSM1676869         GSM1676870         GSM1676871         GSM1676872  \\\n",
       "0   cell line: H2087   cell line: H2087   cell line: H2009   cell line: H2009   \n",
       "1  disease state: --  disease state: --  disease state: --  disease state: --   \n",
       "2  tumor grading: --  tumor grading: --  tumor grading: --  tumor grading: --   \n",
       "3  disease stage: --  disease stage: --  disease stage: --  disease stage: --   \n",
       "4  organism part: --  organism part: --  organism part: --  organism part: --   \n",
       "\n",
       "   ...                              GSM1676991         GSM1676992  \\\n",
       "0  ...                           cell line: --   cell line: HCC78   \n",
       "1  ...  disease state: Squamous Cell Carcinoma  disease state: --   \n",
       "2  ...                       tumor grading: --  tumor grading: --   \n",
       "3  ...                       disease stage: --  disease stage: --   \n",
       "4  ...                     organism part: Lung  organism part: --   \n",
       "\n",
       "          GSM1676993         GSM1676994         GSM1676995         GSM1676996  \\\n",
       "0   cell line: H2347   cell line: H1437   cell line: H2009   cell line: H1437   \n",
       "1  disease state: --  disease state: --  disease state: --  disease state: --   \n",
       "2  tumor grading: --  tumor grading: --  tumor grading: --  tumor grading: --   \n",
       "3  disease stage: --  disease stage: --  disease stage: --  disease stage: --   \n",
       "4  organism part: --  organism part: --  organism part: --  organism part: --   \n",
       "\n",
       "          GSM1676997         GSM1676998         GSM1676999         GSM1677000  \n",
       "0   cell line: H2347   cell line: H2087      cell line: --      cell line: --  \n",
       "1  disease state: --  disease state: --  disease state: --  disease state: --  \n",
       "2  tumor grading: --  tumor grading: --  tumor grading: --  tumor grading: --  \n",
       "3  disease stage: --  disease stage: --  disease stage: --  disease stage: --  \n",
       "4  organism part: --  organism part: --  organism part: --  organism part: --  \n",
       "\n",
       "[5 rows x 138 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['cell line: H2347',\n",
       "  'cell line: H1437',\n",
       "  'cell line: HCC78',\n",
       "  'cell line: H2087',\n",
       "  'cell line: H2009',\n",
       "  'cell line: --'],\n",
       " 1: ['disease state: --',\n",
       "  'disease state: Leiomyoma',\n",
       "  'disease state: Lung_Adenocarcinoma',\n",
       "  'disease state: Conventional_Clear_Cell_Renal_Cell_Carcinoma',\n",
       "  'disease state: Squamous Cell Carcinoma',\n",
       "  'disease state: Stomach Adenocarcinoma',\n",
       "  'disease state: Large Cell Lymphoma',\n",
       "  'disease state: Malignant Melanoma',\n",
       "  'disease state: Recurrent Renal Cell Carcinoma',\n",
       "  'disease state: Adrenal Cortical Adenoma',\n",
       "  'disease state: Ovarian Adenocarcinoma',\n",
       "  'disease state: Gastrointestinal_Stromal_Tumor',\n",
       "  'disease state: Metastatic Renal Cell Carcinoma',\n",
       "  'disease state: Non neoplastic liver with cirrosis',\n",
       "  'disease state: Malignant G1 Stromal Tumor',\n",
       "  'disease state: melanoma'],\n",
       " 2: ['tumor grading: --',\n",
       "  'tumor grading: G2/pT1pN0pMX',\n",
       "  'tumor grading: G3/pT2pN0pMX',\n",
       "  'tumor grading: G2/pT2pN0pMX',\n",
       "  'tumor grading: G3/pT4pNXpMX'],\n",
       " 3: ['disease stage: --',\n",
       "  'disease stage: Stage IA',\n",
       "  'disease stage: Stage IB',\n",
       "  'disease stage: Stage IIIB'],\n",
       " 4: ['organism part: --',\n",
       "  'organism part: Uterus',\n",
       "  'organism part: Lung',\n",
       "  'organism part: Stomach',\n",
       "  'organism part: Lymphoid tissue',\n",
       "  'organism part: Liver',\n",
       "  'organism part: Adrenal Gland',\n",
       "  'organism part: Ovary',\n",
       "  'organism part: Kidney',\n",
       "  'organism part: Skin',\n",
       "  'organism part: Lymph_Node'],\n",
       " 5: ['Sex: --', 'Sex: female', 'Sex: male'],\n",
       " 6: ['age: --', 'age: 67', 'age: 66', 'age: 72', 'age: 56', 'age: 48'],\n",
       " 7: ['histology: --',\n",
       "  'histology: Leiomyoma',\n",
       "  'histology: Lung_Adenocarcinoma',\n",
       "  'histology: Conventional_Clear_Cell_Renal_Cell_Carcinoma',\n",
       "  'histology: Stomach Adenocarcinoma',\n",
       "  'histology: Large Cell Lymphoma',\n",
       "  'histology: Metastatic Malignant Melanoma',\n",
       "  'histology: Recurrent Renal Cell Carcinoma, chromophobe cell type',\n",
       "  'histology: Non neoplastic liver with cirrosis',\n",
       "  'histology: Adrenal Cortical Adenoma',\n",
       "  'histology: Papillary Serous Adenocarcinoma',\n",
       "  'histology: Squamous cell carcinoma 85% tumor 15% Stroma',\n",
       "  'histology: Squamous Cell Carcinoma',\n",
       "  'histology: Malignant G1 Stromal Tumor',\n",
       "  'histology: metastatic renal cell carcinoma',\n",
       "  'histology: Lung Adenocarcinoma',\n",
       "  'histology: carcinoma',\n",
       "  'histology: Adenocarcinoma',\n",
       "  'histology: Squamous Cell carcinoma',\n",
       "  'histology: Metastatic Renal Cell Carcinoma, clear cell type',\n",
       "  'histology: Ovarian Adenocarcinoma',\n",
       "  'histology: Malignant G1 stromal tumor',\n",
       "  'histology: Adenocartcinoma of Lung',\n",
       "  'histology: Squamoous Cell Carcinoma',\n",
       "  'histology: Renal Cell Carcinoma',\n",
       "  'histology: Non neeoplastic liver with cirrosis',\n",
       "  'histology: Metastatic Renal Cell Carcinoma']}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Kidney Chromophobe\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Kidney Chromophobe\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"caArray_dobbi-00100: Interlaboratory comparability study of cancer gene expression analysis using oligonucleotide microarrays\"\\n!Series_summary\\t\"A key step in bringing gene expression data into clinical practice is the conduct of large studies to confirm preliminary models. The performance of such confirmatory studies and the transition to clinical practice requires that microarray data from different laboratories are comparable and reproducible. We designed a study to assess the comparability of data from four laboratories that will conduct a larger microarray profiling confirmation project in lung adenocarcinomas. To test the feasibility of combining data across laboratories, frozen tumor tissues, cell line pellets, and purified RNA samples were analyzed at each of the four laboratories. Samples of each type and several subsamples from each tumor and each cell line were blinded before being distributed. The laboratories followed a common protocol for all steps of tissue processing, RNA extraction, and microarray analysis using Affymetrix Human Genome U133A arrays. High within-laboratory and between-laboratory correlations were observed on the purified RNA samples, the cell lines, and the frozen tumor tissues. Intraclass correlation within laboratories was only slightly stronger than between laboratories, and the intraclass correlation tended to be weakest for genes expressed at low levels and showing small variation. Finally, hierarchical cluster analysis revealed that the repeated samples clustered together regardless of the laboratory in which the experiments were done. The findings indicate that under properly controlled conditions it is feasible to perform complete tumor microarray analysis, from tissue processing to hybridization and scanning, at multiple independent laboratories for a single study.\"\\n!Series_overall_design\\t\"dobbi-00100\"\\n!Series_overall_design\\t\"Assay Type: Gene Expression\"\\n!Series_overall_design\\t\"Provider: Affymetrix\"\\n!Series_overall_design\\t\"Array Designs: HG-U133A\"\\n!Series_overall_design\\t\"Organism: Homo sapiens (ncbitax)\"\\n!Series_overall_design\\t\"Tissue Sites: Kidney, Lung, Stomach, Uterus, Liver, Lymphoid tissue, Ovary, Skin, Adrenal Gland, Lymph_Node\"\\n!Series_overall_design\\t\"Material Types: cell, nuclear_RNA, synthetic_RNA, organism_part, total_RNA\"\\n!Series_overall_design\\t\"Disease States: Recurrent Renal Cell Carcinoma, Squamous Cell Carcinoma,Conventional_Clear_Cell_Renal_Cell_Carcinoma,Gastrointestinal_Stromal_Tumor, Lung_Adenocarcinoma, Leiomyoma, Non neoplastic liver with cirrosis, Stomach Adenocarcinoma, Large Cell Lymphoma, Ovarian Adenocarcinoma, Melanoma, Malignant G1 Stromal Tumor, Adrenal Cortical Adenoma, Metastatic Renal Cell Carcinoma, Malignant Melanoma\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'cell line: H2347\\', \\'cell line: H1437\\', \\'cell line: HCC78\\', \\'cell line: H2087\\', \\'cell line: H2009\\', \\'cell line: --\\'], 1: [\\'disease state: --\\', \\'disease state: Leiomyoma\\', \\'disease state: Lung_Adenocarcinoma\\', \\'disease state: Conventional_Clear_Cell_Renal_Cell_Carcinoma\\', \\'disease state: Squamous Cell Carcinoma\\', \\'disease state: Stomach Adenocarcinoma\\', \\'disease state: Large Cell Lymphoma\\', \\'disease state: Malignant Melanoma\\', \\'disease state: Recurrent Renal Cell Carcinoma\\', \\'disease state: Adrenal Cortical Adenoma\\', \\'disease state: Ovarian Adenocarcinoma\\', \\'disease state: Gastrointestinal_Stromal_Tumor\\', \\'disease state: Metastatic Renal Cell Carcinoma\\', \\'disease state: Non neoplastic liver with cirrosis\\', \\'disease state: Malignant G1 Stromal Tumor\\', \\'disease state: melanoma\\'], 2: [\\'tumor grading: --\\', \\'tumor grading: G2/pT1pN0pMX\\', \\'tumor grading: G3/pT2pN0pMX\\', \\'tumor grading: G2/pT2pN0pMX\\', \\'tumor grading: G3/pT4pNXpMX\\'], 3: [\\'disease stage: --\\', \\'disease stage: Stage IA\\', \\'disease stage: Stage IB\\', \\'disease stage: Stage IIIB\\'], 4: [\\'organism part: --\\', \\'organism part: Uterus\\', \\'organism part: Lung\\', \\'organism part: Stomach\\', \\'organism part: Lymphoid tissue\\', \\'organism part: Liver\\', \\'organism part: Adrenal Gland\\', \\'organism part: Ovary\\', \\'organism part: Kidney\\', \\'organism part: Skin\\', \\'organism part: Lymph_Node\\'], 5: [\\'Sex: --\\', \\'Sex: female\\', \\'Sex: male\\'], 6: [\\'age: --\\', \\'age: 67\\', \\'age: 66\\', \\'age: 72\\', \\'age: 56\\', \\'age: 48\\'], 7: [\\'histology: --\\', \\'histology: Leiomyoma\\', \\'histology: Lung_Adenocarcinoma\\', \\'histology: Conventional_Clear_Cell_Renal_Cell_Carcinoma\\', \\'histology: Stomach Adenocarcinoma\\', \\'histology: Large Cell Lymphoma\\', \\'histology: Metastatic Malignant Melanoma\\', \\'histology: Recurrent Renal Cell Carcinoma, chromophobe cell type\\', \\'histology: Non neoplastic liver with cirrosis\\', \\'histology: Adrenal Cortical Adenoma\\', \\'histology: Papillary Serous Adenocarcinoma\\', \\'histology: Squamous cell carcinoma 85% tumor 15% Stroma\\', \\'histology: Squamous Cell Carcinoma\\', \\'histology: Malignant G1 Stromal Tumor\\', \\'histology: metastatic renal cell carcinoma\\', \\'histology: Lung Adenocarcinoma\\', \\'histology: carcinoma\\', \\'histology: Adenocarcinoma\\', \\'histology: Squamous Cell carcinoma\\', \\'histology: Metastatic Renal Cell Carcinoma, clear cell type\\', \\'histology: Ovarian Adenocarcinoma\\', \\'histology: Malignant G1 stromal tumor\\', \\'histology: Adenocartcinoma of Lung\\', \\'histology: Squamoous Cell Carcinoma\\', \\'histology: Renal Cell Carcinoma\\', \\'histology: Non neeoplastic liver with cirrosis\\', \\'histology: Metastatic Renal Cell Carcinoma\\']}\\n'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 7\n",
    "age_row = 6\n",
    "gender_row = 5\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'histology: Leiomyoma':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if gender_string.lower() == 'sex: male':\n",
    "        return 0\n",
    "    elif gender_string.lower() == 'sex: female':\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1676864</th>\n",
       "      <th>GSM1676865</th>\n",
       "      <th>GSM1676866</th>\n",
       "      <th>GSM1676867</th>\n",
       "      <th>GSM1676868</th>\n",
       "      <th>GSM1676869</th>\n",
       "      <th>GSM1676870</th>\n",
       "      <th>GSM1676871</th>\n",
       "      <th>GSM1676872</th>\n",
       "      <th>GSM1676873</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1676991</th>\n",
       "      <th>GSM1676992</th>\n",
       "      <th>GSM1676993</th>\n",
       "      <th>GSM1676994</th>\n",
       "      <th>GSM1676995</th>\n",
       "      <th>GSM1676996</th>\n",
       "      <th>GSM1676997</th>\n",
       "      <th>GSM1676998</th>\n",
       "      <th>GSM1676999</th>\n",
       "      <th>GSM1677000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GSM1676864 GSM1676865 GSM1676866 GSM1676867 GSM1676868  \\\n",
       "Kidney Chromophobe          0          0          1          0          0   \n",
       "Age                      None       None       None       None       None   \n",
       "Gender                   None       None          1       None       None   \n",
       "\n",
       "                   GSM1676869 GSM1676870 GSM1676871 GSM1676872 GSM1676873  \\\n",
       "Kidney Chromophobe          0          0          0          0          0   \n",
       "Age                      None       None       None       None       None   \n",
       "Gender                   None       None       None       None       None   \n",
       "\n",
       "                    ...  GSM1676991  GSM1676992  GSM1676993  GSM1676994  \\\n",
       "Kidney Chromophobe  ...           0           0           0           0   \n",
       "Age                 ...        None        None        None        None   \n",
       "Gender              ...        None        None        None        None   \n",
       "\n",
       "                    GSM1676995 GSM1676996 GSM1676997 GSM1676998 GSM1676999  \\\n",
       "Kidney Chromophobe           0          0          0          0          0   \n",
       "Age                       None       None       None       None       None   \n",
       "Gender                    None       None       None       None       None   \n",
       "\n",
       "                   GSM1677000  \n",
       "Kidney Chromophobe          0  \n",
       "Age                      None  \n",
       "Gender                   None  \n",
       "\n",
       "[3 rows x 137 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1676864</th>\n",
       "      <th>GSM1676865</th>\n",
       "      <th>GSM1676866</th>\n",
       "      <th>GSM1676867</th>\n",
       "      <th>GSM1676868</th>\n",
       "      <th>GSM1676869</th>\n",
       "      <th>GSM1676870</th>\n",
       "      <th>GSM1676871</th>\n",
       "      <th>GSM1676872</th>\n",
       "      <th>GSM1676873</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1676991</th>\n",
       "      <th>GSM1676992</th>\n",
       "      <th>GSM1676993</th>\n",
       "      <th>GSM1676994</th>\n",
       "      <th>GSM1676995</th>\n",
       "      <th>GSM1676996</th>\n",
       "      <th>GSM1676997</th>\n",
       "      <th>GSM1676998</th>\n",
       "      <th>GSM1676999</th>\n",
       "      <th>GSM1677000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>1932.3800</td>\n",
       "      <td>1032.6500</td>\n",
       "      <td>1282.8000</td>\n",
       "      <td>2688.6100</td>\n",
       "      <td>2189.2700</td>\n",
       "      <td>342.6390</td>\n",
       "      <td>254.9960</td>\n",
       "      <td>2225.930</td>\n",
       "      <td>2785.7000</td>\n",
       "      <td>1785.5800</td>\n",
       "      <td>...</td>\n",
       "      <td>2212.0200</td>\n",
       "      <td>1105.370</td>\n",
       "      <td>1429.44000</td>\n",
       "      <td>1034.43000</td>\n",
       "      <td>1963.0400</td>\n",
       "      <td>685.8430</td>\n",
       "      <td>1038.4300</td>\n",
       "      <td>267.5860</td>\n",
       "      <td>854.4560</td>\n",
       "      <td>1051.0300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>833.8050</td>\n",
       "      <td>1034.2300</td>\n",
       "      <td>647.4520</td>\n",
       "      <td>149.0560</td>\n",
       "      <td>315.9340</td>\n",
       "      <td>155.5640</td>\n",
       "      <td>312.1450</td>\n",
       "      <td>431.223</td>\n",
       "      <td>325.0920</td>\n",
       "      <td>793.2160</td>\n",
       "      <td>...</td>\n",
       "      <td>199.2370</td>\n",
       "      <td>514.352</td>\n",
       "      <td>879.66300</td>\n",
       "      <td>1314.85000</td>\n",
       "      <td>625.2920</td>\n",
       "      <td>727.3050</td>\n",
       "      <td>521.0290</td>\n",
       "      <td>205.0130</td>\n",
       "      <td>478.0530</td>\n",
       "      <td>466.5580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>122.2550</td>\n",
       "      <td>59.3265</td>\n",
       "      <td>126.7500</td>\n",
       "      <td>139.8990</td>\n",
       "      <td>98.9554</td>\n",
       "      <td>66.8850</td>\n",
       "      <td>74.1553</td>\n",
       "      <td>131.672</td>\n",
       "      <td>77.3771</td>\n",
       "      <td>120.1670</td>\n",
       "      <td>...</td>\n",
       "      <td>429.3040</td>\n",
       "      <td>101.457</td>\n",
       "      <td>91.43080</td>\n",
       "      <td>83.34430</td>\n",
       "      <td>197.2620</td>\n",
       "      <td>115.2990</td>\n",
       "      <td>128.3080</td>\n",
       "      <td>84.8303</td>\n",
       "      <td>110.7950</td>\n",
       "      <td>94.8872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>1134.5400</td>\n",
       "      <td>1058.9700</td>\n",
       "      <td>1107.4800</td>\n",
       "      <td>1712.0000</td>\n",
       "      <td>1175.1700</td>\n",
       "      <td>1004.8500</td>\n",
       "      <td>943.4220</td>\n",
       "      <td>1246.450</td>\n",
       "      <td>1114.2500</td>\n",
       "      <td>1289.7300</td>\n",
       "      <td>...</td>\n",
       "      <td>1237.0500</td>\n",
       "      <td>848.331</td>\n",
       "      <td>760.27900</td>\n",
       "      <td>608.74700</td>\n",
       "      <td>750.6020</td>\n",
       "      <td>815.2660</td>\n",
       "      <td>822.6670</td>\n",
       "      <td>513.1220</td>\n",
       "      <td>741.4220</td>\n",
       "      <td>630.6230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>26.6585</td>\n",
       "      <td>41.4460</td>\n",
       "      <td>58.2286</td>\n",
       "      <td>77.8601</td>\n",
       "      <td>70.6387</td>\n",
       "      <td>37.9141</td>\n",
       "      <td>16.5508</td>\n",
       "      <td>106.737</td>\n",
       "      <td>77.7958</td>\n",
       "      <td>67.8281</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0356</td>\n",
       "      <td>16.150</td>\n",
       "      <td>8.93552</td>\n",
       "      <td>3.17154</td>\n",
       "      <td>93.9857</td>\n",
       "      <td>33.7936</td>\n",
       "      <td>25.4411</td>\n",
       "      <td>25.6873</td>\n",
       "      <td>32.2491</td>\n",
       "      <td>12.3147</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 137 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM1676864  GSM1676865  GSM1676866  GSM1676867  GSM1676868  \\\n",
       "ID                                                                      \n",
       "1007_s_at   1932.3800   1032.6500   1282.8000   2688.6100   2189.2700   \n",
       "1053_at      833.8050   1034.2300    647.4520    149.0560    315.9340   \n",
       "117_at       122.2550     59.3265    126.7500    139.8990     98.9554   \n",
       "121_at      1134.5400   1058.9700   1107.4800   1712.0000   1175.1700   \n",
       "1255_g_at     26.6585     41.4460     58.2286     77.8601     70.6387   \n",
       "\n",
       "           GSM1676869  GSM1676870  GSM1676871  GSM1676872  GSM1676873  ...  \\\n",
       "ID                                                                     ...   \n",
       "1007_s_at    342.6390    254.9960    2225.930   2785.7000   1785.5800  ...   \n",
       "1053_at      155.5640    312.1450     431.223    325.0920    793.2160  ...   \n",
       "117_at        66.8850     74.1553     131.672     77.3771    120.1670  ...   \n",
       "121_at      1004.8500    943.4220    1246.450   1114.2500   1289.7300  ...   \n",
       "1255_g_at     37.9141     16.5508     106.737     77.7958     67.8281  ...   \n",
       "\n",
       "           GSM1676991  GSM1676992  GSM1676993  GSM1676994  GSM1676995  \\\n",
       "ID                                                                      \n",
       "1007_s_at   2212.0200    1105.370  1429.44000  1034.43000   1963.0400   \n",
       "1053_at      199.2370     514.352   879.66300  1314.85000    625.2920   \n",
       "117_at       429.3040     101.457    91.43080    83.34430    197.2620   \n",
       "121_at      1237.0500     848.331   760.27900   608.74700    750.6020   \n",
       "1255_g_at     19.0356      16.150     8.93552     3.17154     93.9857   \n",
       "\n",
       "           GSM1676996  GSM1676997  GSM1676998  GSM1676999  GSM1677000  \n",
       "ID                                                                     \n",
       "1007_s_at    685.8430   1038.4300    267.5860    854.4560   1051.0300  \n",
       "1053_at      727.3050    521.0290    205.0130    478.0530    466.5580  \n",
       "117_at       115.2990    128.3080     84.8303    110.7950     94.8872  \n",
       "121_at       815.2660    822.6670    513.1220    741.4220    630.6230  \n",
       "1255_g_at     33.7936     25.4411     25.6873     32.2491     12.3147  \n",
       "\n",
       "[5 rows x 137 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1007_s_at',\n",
       " '1053_at',\n",
       " '117_at',\n",
       " '121_at',\n",
       " '1255_g_at',\n",
       " '1294_at',\n",
       " '1316_at',\n",
       " '1320_at',\n",
       " '1405_i_at',\n",
       " '1431_at',\n",
       " '1438_at',\n",
       " '1487_at',\n",
       " '1494_f_at',\n",
       " '1598_g_at',\n",
       " '160020_at',\n",
       " '1729_at',\n",
       " '1773_at',\n",
       " '177_at',\n",
       " '179_at',\n",
       " '1861_at']"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1598_g_at', '160020_at', '1729_at', '1773_at', '177_at', '179_at', '1861_at']\\n\""
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'UCSC_RefGene_Name'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_a\n",
      "1000 input query terms found no hit:\t['201473_at', '201474_s_at', '201475_x_at', '201476_s_at', '201477_s_at', '201478_s_at', '201479_at'\n",
      "1000 input query terms found no hit:\t['202473_x_at', '202474_s_at', '202475_at', '202476_s_at', '202477_s_at', '202478_at', '202479_s_at'\n",
      "1000 input query terms found no hit:\t['203474_at', '203475_at', '203476_at', '203477_at', '203478_at', '203479_s_at', '203480_s_at', '203\n",
      "1000 input query terms found no hit:\t['204474_at', '204475_at', '204476_s_at', '204477_at', '204478_s_at', '204479_at', '204480_s_at', '2\n",
      "1000 input query terms found no hit:\t['205474_at', '205475_at', '205476_at', '205477_s_at', '205478_at', '205479_s_at', '205480_s_at', '2\n",
      "1000 input query terms found no hit:\t['206475_x_at', '206476_s_at', '206477_s_at', '206478_at', '206479_at', '206480_at', '206481_s_at', \n",
      "1000 input query terms found no hit:\t['207479_at', '207480_s_at', '207481_at', '207482_at', '207483_s_at', '207484_s_at', '207485_x_at', \n",
      "1000 input query terms found no hit:\t['208503_s_at', '208504_x_at', '208505_s_at', '208506_at', '208507_at', '208508_s_at', '208509_s_at'\n",
      "1000 input query terms found no hit:\t['209508_x_at', '209509_s_at', '209510_at', '209511_at', '209512_at', '209513_s_at', '209514_s_at', \n",
      "1000 input query terms found no hit:\t['210533_at', '210534_s_at', '210535_at', '210536_s_at', '210537_s_at', '210538_s_at', '210539_at', \n",
      "1000 input query terms found no hit:\t['211596_s_at', '211597_s_at', '211598_x_at', '211599_x_at', '211600_at', '211601_at', '211602_s_at'\n",
      "1000 input query terms found no hit:\t['212616_at', '212617_at', '212618_at', '212619_at', '212620_at', '212621_at', '212622_at', '212623_\n",
      "1000 input query terms found no hit:\t['213621_s_at', '213622_at', '213623_at', '213624_at', '213625_at', '213626_at', '213627_at', '21362\n",
      "1000 input query terms found no hit:\t['214623_at', '214624_at', '214625_s_at', '214626_s_at', '214627_at', '214628_at', '214629_x_at', '2\n",
      "1000 input query terms found no hit:\t['215628_x_at', '215629_s_at', '215630_at', '215631_s_at', '215632_at', '215633_x_at', '215634_at', \n",
      "1000 input query terms found no hit:\t['216631_s_at', '216632_at', '216633_s_at', '216634_at', '216635_at', '216636_at', '216637_at', '216\n",
      "1000 input query terms found no hit:\t['217636_at', '217637_at', '217638_at', '217639_at', '217640_x_at', '217641_at', '217642_at', '21764\n",
      "1000 input query terms found no hit:\t['218637_at', '218638_s_at', '218639_s_at', '218640_s_at', '218641_at', '218642_s_at', '218643_s_at'\n",
      "1000 input query terms found no hit:\t['219637_at', '219638_at', '219639_x_at', '219640_at', '219641_at', '219642_s_at', '219643_at', '219\n",
      "1000 input query terms found no hit:\t['220637_at', '220638_s_at', '220639_at', '220640_at', '220641_at', '220642_x_at', '220643_s_at', '2\n",
      "1000 input query terms found no hit:\t['221640_s_at', '221641_s_at', '221642_at', '221643_s_at', '221644_s_at', '221645_s_at', '221646_s_a\n",
      "283 input query terms found no hit:\t['44146_at', '44563_at', '44617_at', '44654_at', '44669_at', '44673_at', '44696_at', '44702_at', '44\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 20 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Kidney Chromophobe', the least common label is '0.0' with 20 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Kidney Chromophobe' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 56.0\n",
      "  50% (Median): 66.0\n",
      "  75%: 67.0\n",
      "Min: 48.0\n",
      "Max: 72.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 8 occurrences. This represents 40.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xena</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>91</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "0      Xena       True          True      False     True        True   \n",
       "\n",
       "   sample_size note  \n",
       "0           91       "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Xena'"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xena</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>91</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "0      Xena       True          True      False     True        True   \n",
       "\n",
       "   sample_size note  \n",
       "0           91       "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.295070023Z",
     "start_time": "2024-01-10T21:40:21.338515691Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kidney Chromophobe</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.295492</td>\n",
       "      <td>-3.487426</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.359428</td>\n",
       "      <td>0.850422</td>\n",
       "      <td>0.58019</td>\n",
       "      <td>-0.739994</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>0.397618</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>0.731583</td>\n",
       "      <td>-1.308333</td>\n",
       "      <td>-0.79709</td>\n",
       "      <td>-1.328077</td>\n",
       "      <td>0.933073</td>\n",
       "      <td>2.015005</td>\n",
       "      <td>0.025067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>0.581408</td>\n",
       "      <td>0.368474</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.217628</td>\n",
       "      <td>0.626922</td>\n",
       "      <td>0.06679</td>\n",
       "      <td>-0.058894</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.193178</td>\n",
       "      <td>0.513683</td>\n",
       "      <td>-0.912933</td>\n",
       "      <td>-0.32549</td>\n",
       "      <td>-1.035777</td>\n",
       "      <td>0.221073</td>\n",
       "      <td>1.600605</td>\n",
       "      <td>2.456767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>1.119008</td>\n",
       "      <td>2.198374</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.341628</td>\n",
       "      <td>0.681022</td>\n",
       "      <td>0.46319</td>\n",
       "      <td>0.057806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.837786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.006622</td>\n",
       "      <td>5.587783</td>\n",
       "      <td>0.113667</td>\n",
       "      <td>-0.03259</td>\n",
       "      <td>0.051323</td>\n",
       "      <td>1.403473</td>\n",
       "      <td>0.786105</td>\n",
       "      <td>3.303867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>0.572008</td>\n",
       "      <td>-1.889926</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.683828</td>\n",
       "      <td>0.971922</td>\n",
       "      <td>1.55029</td>\n",
       "      <td>-0.150194</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>0.414383</td>\n",
       "      <td>-2.250633</td>\n",
       "      <td>1.13531</td>\n",
       "      <td>-1.257677</td>\n",
       "      <td>-1.076027</td>\n",
       "      <td>1.424005</td>\n",
       "      <td>-0.464633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>0.259208</td>\n",
       "      <td>-0.380726</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.992728</td>\n",
       "      <td>0.311022</td>\n",
       "      <td>0.65699</td>\n",
       "      <td>0.050806</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.295086</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>1.982883</td>\n",
       "      <td>-0.845733</td>\n",
       "      <td>1.40521</td>\n",
       "      <td>-1.536277</td>\n",
       "      <td>0.100373</td>\n",
       "      <td>1.883005</td>\n",
       "      <td>0.013067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Kidney Chromophobe  Age  Gender  ARHGEF10L     HIF3A     RNF17     RNF10  \\\n",
       "0                   1   57       0  -0.295492 -3.487426 -0.531035  0.359428   \n",
       "1                   1   67       0   0.581408  0.368474 -0.531035  1.217628   \n",
       "2                   0   67       0   1.119008  2.198374 -0.531035  0.341628   \n",
       "3                   1   56       0   0.572008 -1.889926 -0.531035  0.683828   \n",
       "4                   1   69       1   0.259208 -0.380726 -0.531035  0.992728   \n",
       "\n",
       "      RNF11    RNF13   GTF2IP1  ...   SLC7A10   PLA2G2C     TULP2     NPY5R  \\\n",
       "0  0.850422  0.58019 -0.739994  ... -2.090786  0.397618 -0.748878  0.731583   \n",
       "1  0.626922  0.06679 -0.058894  ... -2.090786 -0.086682 -0.193178  0.513683   \n",
       "2  0.681022  0.46319  0.057806  ... -0.837786 -0.086682  0.006622  5.587783   \n",
       "3  0.971922  1.55029 -0.150194  ... -2.090786 -0.086682 -0.748878  0.414383   \n",
       "4  0.311022  0.65699  0.050806  ... -0.295086 -0.086682 -0.748878  1.982883   \n",
       "\n",
       "      GNGT2    GNGT1     TULP3     BCL6B     GSTK1      SELP  \n",
       "0 -1.308333 -0.79709 -1.328077  0.933073  2.015005  0.025067  \n",
       "1 -0.912933 -0.32549 -1.035777  0.221073  1.600605  2.456767  \n",
       "2  0.113667 -0.03259  0.051323  1.403473  0.786105  3.303867  \n",
       "3 -2.250633  1.13531 -1.257677 -1.076027  1.424005 -0.464633  \n",
       "4 -0.845733  1.40521 -1.536277  0.100373  1.883005  0.013067  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.358779256Z",
     "start_time": "2024-01-10T21:40:27.271337819Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n",
    "\n",
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_batch_effect(X):\n",
    "    \"\"\"\n",
    "    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n",
    "\n",
    "    Args:\n",
    "    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n",
    "\n",
    "    Returns:\n",
    "    bool: True if a potential batch effect is detected, False otherwise.\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "\n",
    "    # Computing XX^T\n",
    "    XXt = np.dot(X, X.T)\n",
    "\n",
    "    # Compute the eigenvalues of XX^T\n",
    "    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n",
    "    eigen_values = sorted(eigen_values, reverse=True)\n",
    "\n",
    "    # Check for large gaps in the eigenvalues\n",
    "    for i in range(len(eigen_values) - 1):\n",
    "        gap = eigen_values[i] - eigen_values[i + 1]\n",
    "        if gap > 1 / n_samples:  # You may need to adjust this threshold\n",
    "            return True\n",
    "\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546590575Z",
     "start_time": "2024-01-10T21:40:27.298495045Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sparse_lmm import VariableSelection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546733889Z",
     "start_time": "2024-01-10T21:40:27.535249532Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Select appropriate models based on whether the dataset has batch effect.\n",
    "# We experiment on two models for each branch. We will decide which one to choose later.\n",
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n",
    "    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n",
    "    indices = np.arange(X.shape[0])\n",
    "    np.random.shuffle(indices)\n",
    "\n",
    "    fold_size = len(X) // k\n",
    "    performances = []\n",
    "\n",
    "    for i in range(k):\n",
    "        # Split data into train and test based on the current fold\n",
    "        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n",
    "        train_indices = np.setdiff1d(indices, test_indices)\n",
    "\n",
    "        X_train, X_test = X[train_indices], X[test_indices]\n",
    "        Y_train, Y_test = Y[train_indices], Y[test_indices]\n",
    "        Z_train, Z_test = Z[train_indices], Z[test_indices]\n",
    "\n",
    "        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n",
    "        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n",
    "\n",
    "        # model = model_constructor(**model_params)\n",
    "        model = ResidualizationRegressor(model_constructor, model_params)\n",
    "        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n",
    "        predictions = model.predict(normalized_X_test, normalized_Z_test)\n",
    "\n",
    "        if target_type == 'binary':\n",
    "            predictions = (predictions > 0.5).astype(int)\n",
    "            Y_test = (Y_test > 0.5).astype(int)\n",
    "            performance = accuracy_score(Y_test, predictions)\n",
    "        elif target_type == 'continuous':\n",
    "            performance = mean_squared_error(Y_test, predictions)\n",
    "\n",
    "        performances.append(performance)\n",
    "\n",
    "    cv_mean = np.mean(performances)\n",
    "    cv_std = np.std(performances)\n",
    "\n",
    "    if target_type == 'binary':\n",
    "        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n",
    "    else:\n",
    "        print(f'The cross-validation MSE is {(cv_mean * 100):.2f} ± {(cv_std * 100):.2f}')\n",
    "\n",
    "    return cv_mean, cv_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_data(X_train, X_test=None):\n",
    "    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n",
    "    and optionally the test data\"\"\"\n",
    "    mean = np.mean(X_train, axis=0)\n",
    "    std = np.std(X_train, axis=0)\n",
    "\n",
    "    # Handling columns with std = 0\n",
    "    std_no_zero = np.where(std == 0, 1, std)\n",
    "\n",
    "    # Normalize X_train\n",
    "    X_train_normalized = (X_train - mean) / std_no_zero\n",
    "    # Set normalized values to 0 where std was 0\n",
    "    X_train_normalized[:, std == 0] = 0\n",
    "\n",
    "    if X_test is not None:\n",
    "        X_test_normalized = (X_test - mean) / std_no_zero\n",
    "        X_test_normalized[:, std == 0] = 0\n",
    "    else:\n",
    "        X_test_normalized = None\n",
    "\n",
    "    return X_train_normalized, X_test_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ResidualizationRegressor:\n",
    "    def __init__(self, regression_model_constructor, params=None):\n",
    "        if params is None:\n",
    "            params = {}\n",
    "        self.regression_model = regression_model_constructor(**params)\n",
    "        self.beta_Z = None  # Coefficients for regression of Y on Z\n",
    "        self.beta_X = None  # Coefficients for regression of residual on X\n",
    "        self.neg_log_p_values = None  # Negative logarithm of p-values\n",
    "        self.p_values = None  # Actual p-values\n",
    "\n",
    "    def _reshape_data(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the data to ensure it's in the correct format (2D array).\n",
    "\n",
    "        :param data: The input data (can be 1D or 2D array).\n",
    "        :return: Reshaped 2D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 1:\n",
    "            return data.reshape(-1, 1)\n",
    "        return data\n",
    "\n",
    "    def _reshape_output(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the output data to ensure it's in the correct format (1D array).\n",
    "\n",
    "        :param data: The output data (can be 1D or 2D array).\n",
    "        :return: Reshaped 1D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 2 and data.shape[1] == 1:\n",
    "            return data.ravel()\n",
    "        return data\n",
    "    \n",
    "    def fit(self, X, Y, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Y = self._reshape_data(Y)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        # Step 1: Linear regression of Y on Z\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n",
    "        Y_hat = Z_ones @ self.beta_Z\n",
    "        e_Y = Y - Y_hat  # Residual of Y\n",
    "\n",
    "        # Step 2: Regress the residual on X using the included regression model\n",
    "        self.regression_model.fit(X, e_Y)\n",
    "\n",
    "        # Obtain coefficients from the regression model\n",
    "        if hasattr(self.regression_model, 'coef_'):\n",
    "            self.beta_X = self.regression_model.coef_\n",
    "        elif hasattr(self.regression_model, 'getBeta'):\n",
    "            beta_output = self.regression_model.getBeta()\n",
    "            self.beta_X = self._reshape_output(beta_output)\n",
    "\n",
    "        # Obtain negative logarithm of p-values, if available\n",
    "        if hasattr(self.regression_model, 'getNegLogP'):\n",
    "            neg_log_p_output = self.regression_model.getNegLogP()\n",
    "            if neg_log_p_output is not None:\n",
    "                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n",
    "                self.p_values = np.exp(-self.neg_log_p_values)\n",
    "                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n",
    "                p_values_Z = np.full(Z.shape[1], np.nan)\n",
    "                self.p_values = np.concatenate((p_values_Z, self.p_values))\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, mean_squared_error\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:30.570471636Z",
     "start_time": "2024-01-10T21:40:27.546295159Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "The cross-validation accuracy is 64.44% ± 15.56%\n"
     ]
    }
   ],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:47.575878022Z",
     "start_time": "2024-01-10T21:40:30.570907992Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n",
      "  ts = beta / np.sqrt(var * sigma)\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n",
      "  ts = beta / np.sqrt(var * sigma)\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n",
      "  ts = beta / np.sqrt(var * sigma)\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n",
      "  ts = beta / np.sqrt(var * sigma)\n",
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n",
      "  ts = beta / np.sqrt(var * sigma)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The cross-validation accuracy is 96.67% ± 4.44%\n"
     ]
    }
   ],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.663737008Z",
     "start_time": "2024-01-10T21:40:47.576021186Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha for Lasso: 0.0003\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n",
      "  ts = beta / np.sqrt(var * sigma)\n"
     ]
    }
   ],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n",
    "                     threshold: float = 0.05, save_output: bool = True,\n",
    "                     output_dir: str = './output', model_id: int = 1) -> None:\n",
    "    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n",
    "    consists of one variable about condition and multiple variables about genetic factors.\n",
    "    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n",
    "    p-values or non-zero coefficients, depending on the availability of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    model (Any): The trained regression Model.\n",
    "    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n",
    "    trait (str): The target trait of interest.\n",
    "    condition (str): The specific condition to examine within the model.\n",
    "    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n",
    "    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n",
    "    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n",
    "    model_id (int): The index of the model, 1 or 2.\n",
    "\n",
    "    Returns:\n",
    "    None: This function does not return anything but prints and optionally saves the output.\n",
    "    \"\"\"\n",
    "    coefficients = model.get_coefficients().reshape(-1).tolist()\n",
    "    p_values = model.get_p_values()\n",
    "    if p_values is None:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients\n",
    "        })\n",
    "    else:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients,\n",
    "            'p_value': p_values.reshape(-1).tolist()\n",
    "        })\n",
    "    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n",
    "\n",
    "    print(f\"Effect of the condition on the target variable:\")\n",
    "    print(f\"Variable: {condition}\")\n",
    "    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n",
    "    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n",
    "    if p_values is None:\n",
    "        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n",
    "        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n",
    "        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n",
    "            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n",
    "    else:\n",
    "        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n",
    "        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n",
    "        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n",
    "        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n",
    "        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n",
    "            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n",
    "\n",
    "    print(significant_genes_sorted.to_string(index=False))\n",
    "\n",
    "    # Optionally, save this to a CSV file\n",
    "    if save_output:\n",
    "        significant_genes_sorted.to_csv(\n",
    "            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.673565176Z",
     "start_time": "2024-01-10T21:40:51.665291317Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect of the condition on the target variable:\n",
      "Variable: Age\n",
      "Coefficient: -0.0364\n",
      "Found 75 genes with non-zero coefficients associated with the trait 'Kidney Chromophobe' conditional on the factor 'Age'. These genes are identified as significant based on the regression model.\n",
      " Variable  Coefficient  Absolute Coefficient\n",
      "    FOXF2    -6.475829              6.475829\n",
      "    TAF1D    -6.007020              6.007020\n",
      "  SLC30A2     4.455768              4.455768\n",
      "   OR51A4    -4.380482              4.380482\n",
      "     PID1    -3.880443              3.880443\n",
      "  PGLYRP2    -3.482427              3.482427\n",
      "   LRSAM1     3.342719              3.342719\n",
      "    GJA10    -3.283148              3.283148\n",
      "    GALR2     3.246703              3.246703\n",
      "    MEX3B    -2.954832              2.954832\n",
      "     LBX1    -2.931299              2.931299\n",
      "     VAPA    -2.849938              2.849938\n",
      "     IRS1    -2.795838              2.795838\n",
      " FAM41AY1     2.554983              2.554983\n",
      "    SEH1L    -2.442089              2.442089\n",
      "    FOXB1    -2.425475              2.425475\n",
      " ALDH16A1     2.365099              2.365099\n",
      "    TRPC6    -2.328060              2.328060\n",
      "  KIR3DL3     2.270958              2.270958\n",
      "     DPP7     2.092924              2.092924\n",
      "   CHI3L1    -1.849111              1.849111\n",
      "   BCKDHB    -1.845764              1.845764\n",
      "  PGLYRP4    -1.727673              1.727673\n",
      "    MARK1    -1.721585              1.721585\n",
      "    RAB6C    -1.711587              1.711587\n",
      "    SEBOX     1.562871              1.562871\n",
      "   SNORA9    -1.555408              1.555408\n",
      "   GUSBP3    -1.493738              1.493738\n",
      "   TAS1R2    -1.456911              1.456911\n",
      "  B4GALT6    -1.437249              1.437249\n",
      "   ZNF596     1.430365              1.430365\n",
      "    CFHR4     1.428949              1.428949\n",
      "     MMD2    -1.376524              1.376524\n",
      "     JPH1    -1.372318              1.372318\n",
      "     EVX1    -1.364811              1.364811\n",
      "    PCSK1    -1.361661              1.361661\n",
      "   TRMT11    -1.296436              1.296436\n",
      "    AMELY     1.264743              1.264743\n",
      "   RBMXL3    -1.263035              1.263035\n",
      "    KRT6A    -1.232592              1.232592\n",
      "   MMP23A     1.198870              1.198870\n",
      "   OR2T35    -1.192327              1.192327\n",
      "   KCNK18    -1.172123              1.172123\n",
      "   CACNG5    -1.107035              1.107035\n",
      "  CCDC90B    -1.099112              1.099112\n",
      "   GRXCR2     1.089371              1.089371\n",
      "    FANCL    -1.018960              1.018960\n",
      "      MMD    -0.990490              0.990490\n",
      "  GPR137C    -0.980506              0.980506\n",
      "    EHMT1     0.911376              0.911376\n",
      "  SNORA47    -0.863551              0.863551\n",
      "LOC729609    -0.859135              0.859135\n",
      "     HBA1     0.691083              0.691083\n",
      "     XKR9    -0.603696              0.603696\n",
      "   ZNF836    -0.602130              0.602130\n",
      "    SPRR3    -0.559968              0.559968\n",
      "   OR5AN1    -0.555841              0.555841\n",
      "   CLEC4M     0.516088              0.516088\n",
      "   PMCHL2     0.447473              0.447473\n",
      "    OR5P2    -0.433324              0.433324\n",
      "    TUBB8     0.389524              0.389524\n",
      "  GPR37L1     0.386678              0.386678\n",
      "     SGTB    -0.239226              0.239226\n",
      "     TPTE    -0.215320              0.215320\n",
      "     LY6D    -0.180589              0.180589\n",
      "     GDF2    -0.164557              0.164557\n",
      "   AMZ2P1    -0.156981              0.156981\n",
      " SNORA14B    -0.125795              0.125795\n",
      "    SIDT1     0.117917              0.117917\n",
      "    RPH3A     0.077393              0.077393\n",
      " C11orf42    -0.076580              0.076580\n",
      "   TMEM40    -0.076126              0.076126\n",
      "    ARL5C    -0.045519              0.045519\n",
      "   MAGEB4    -0.019771              0.019771\n",
      "    IFNA4    -0.006015              0.006015\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/hx/qdnjfcxj2kbfy5k46ywrbrz40000gn/T/ipykernel_4203/2152198575.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n"
     ]
    }
   ],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.730825913Z",
     "start_time": "2024-01-10T21:40:51.675049728Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect of the condition on the target variable:\n",
      "Variable: Age\n",
      "Coefficient: -0.0364\n",
      "Found 197 significant genes associated with the trait 'Kidney Chromophobe' conditional on the factor 'Age', with corrected p-value < 0.05:\n",
      "Variable  Coefficient      p_value  corrected_p_value\n",
      "  UGT2A3    -0.285658 5.341910e-15       4.750530e-11\n",
      "   RALYL    -0.309340 5.479589e-15       4.750530e-11\n",
      "  PTGER1    -0.275027 5.602681e-13       3.238163e-09\n",
      "  UGT3A1    -0.240755 1.502332e-12       6.512235e-09\n",
      "  MAPK15    -0.194648 1.964597e-12       6.812828e-09\n",
      "  GABRA2    -0.253559 3.741700e-11       1.081289e-07\n",
      "  CLDN19    -0.222469 1.784410e-09       3.437764e-06\n",
      "    IRX1    -0.210113 1.665566e-09       3.437764e-06\n",
      "    FUT6    -0.205566 1.655916e-09       3.437764e-06\n",
      "    VIL1    -0.194274 2.204524e-09       3.822423e-06\n",
      "   MORN5    -0.202149 3.074520e-09       4.846281e-06\n",
      "   FOXJ1    -0.234478 3.442289e-09       4.973820e-06\n",
      "   UPK1B    -0.180007 5.028668e-09       6.707082e-06\n",
      "  SLC9A3    -0.205328 5.600647e-09       6.936401e-06\n",
      "   FRMD1    -0.190063 9.520921e-09       1.100555e-05\n",
      "  FIGNL2    -0.184452 1.482120e-08       1.606155e-05\n",
      "    EDAR    -0.167143 1.777543e-08       1.712268e-05\n",
      "   TNNT2    -0.172218 1.691307e-08       1.712268e-05\n",
      "    NAT2    -0.173341 3.379720e-08       3.084261e-05\n",
      "   TEKT2    -0.190460 3.895283e-08       3.377016e-05\n",
      "    TSG1    -0.161566 6.525130e-08       5.235129e-05\n",
      "    IRX2    -0.181697 6.642415e-08       5.235129e-05\n",
      "    MYH8    -0.193565 1.003007e-07       7.561364e-05\n",
      "  TSPAN8    -0.166418 1.304871e-07       9.427148e-05\n",
      "   HAGHL     0.205804 1.475335e-07       1.023234e-04\n",
      "   VGLL1    -0.186966 1.554776e-07       1.036856e-04\n",
      "  ACSM2A    -0.162882 1.885308e-07       1.210717e-04\n",
      "    FTCD    -0.143408 2.891259e-07       1.790412e-04\n",
      "   CNTN5    -0.138988 3.258719e-07       1.948377e-04\n",
      "  TRIM15    -0.153281 3.428719e-07       1.981685e-04\n",
      "    GBA3    -0.158281 3.817528e-07       2.135230e-04\n",
      "   CAPSL    -0.123800 4.093819e-07       2.218210e-04\n",
      "   NAPSA    -0.177706 5.520656e-07       2.900687e-04\n",
      "   ARMC3    -0.129264 6.123311e-07       3.122709e-04\n",
      "    ZBBX    -0.124271 7.534481e-07       3.628899e-04\n",
      " DYNLRB2    -0.169330 7.481943e-07       3.628899e-04\n",
      " SLC23A1    -0.156575 9.020234e-07       4.227077e-04\n",
      "   ALPK2    -0.147038 9.949225e-07       4.423323e-04\n",
      "   ABCA4    -0.163780 9.876219e-07       4.423323e-04\n",
      "  CLDN16    -0.148924 1.040865e-06       4.511888e-04\n",
      "   ENKUR    -0.123884 1.135708e-06       4.688583e-04\n",
      " SLC13A1    -0.151242 1.119058e-06       4.688583e-04\n",
      "  CYP2B6    -0.158145 1.472542e-06       5.937770e-04\n",
      "  ACSM2B    -0.157747 1.513474e-06       5.964120e-04\n",
      " C2orf73    -0.175334 1.781120e-06       6.570815e-04\n",
      "   FGFR3    -0.167575 1.735003e-06       6.570815e-04\n",
      " IL22RA1    -0.133469 1.751541e-06       6.570815e-04\n",
      "  FAM81B    -0.158515 1.834667e-06       6.627351e-04\n",
      "  TRIM10    -0.155208 2.137495e-06       7.563679e-04\n",
      "   CALB1    -0.116185 3.251595e-06       1.127588e-03\n",
      "    AQP2    -0.149412 3.518400e-06       1.196187e-03\n",
      " SLC44A4    -0.113592 3.605391e-06       1.202190e-03\n",
      "  ANXA13    -0.108037 4.015589e-06       1.313704e-03\n",
      "    RBP4    -0.146457 4.109494e-06       1.319528e-03\n",
      "   CDHR5    -0.145775 4.195635e-06       1.322693e-03\n",
      "  SPIRE1     0.141435 4.442537e-06       1.375521e-03\n",
      "    LYG1    -0.155648 4.921891e-06       1.494861e-03\n",
      " C9orf50    -0.149746 5.000400e-06       1.494861e-03\n",
      " SLC17A4    -0.124892 5.110753e-06       1.501955e-03\n",
      "  ANKS4B    -0.141686 5.241247e-06       1.514633e-03\n",
      " SLC12A1    -0.179747 6.312346e-06       1.777913e-03\n",
      "   TYRP1    -0.160530 6.562458e-06       1.777913e-03\n",
      "   PF4V1    -0.102488 6.391148e-06       1.777913e-03\n",
      "   GLYAT    -0.147979 6.530432e-06       1.777913e-03\n",
      "    A1CF    -0.142365 7.407591e-06       1.976003e-03\n",
      "   DACH2    -0.132060 7.891368e-06       2.073158e-03\n",
      "   DACT2    -0.154978 9.677873e-06       2.504547e-03\n",
      "    THPO    -0.123534 1.000918e-05       2.552193e-03\n",
      "   CLRN3    -0.129456 1.077778e-05       2.708348e-03\n",
      "   KLRG2    -0.131094 1.140998e-05       2.826253e-03\n",
      " ZMYND10    -0.109225 1.209283e-05       2.953205e-03\n",
      "SLC22A11    -0.125119 1.249331e-05       3.008632e-03\n",
      "    UNCX    -0.137284 1.429828e-05       3.396136e-03\n",
      "    KLK4     0.122850 1.533386e-05       3.548634e-03\n",
      "  GPRIN2    -0.132660 1.534965e-05       3.548634e-03\n",
      "  PRODH2    -0.129660 1.604226e-05       3.659958e-03\n",
      "   DPPA2    -0.070013 1.791760e-05       4.034717e-03\n",
      "   PHKA1     0.143014 1.940195e-05       4.312954e-03\n",
      "    PROC    -0.134285 2.009226e-05       4.409869e-03\n",
      "    DPYS    -0.131016 2.228489e-05       4.770342e-03\n",
      "   DCDC1    -0.121219 2.206239e-05       4.770342e-03\n",
      " SLCO1A2    -0.138533 2.769869e-05       5.854467e-03\n",
      "SERPINA6    -0.118861 2.802473e-05       5.854467e-03\n",
      "   NXNL2    -0.107423 2.903353e-05       5.993005e-03\n",
      "  CYP8B1    -0.096923 3.010271e-05       6.140598e-03\n",
      " TMEM207    -0.142187 3.180999e-05       6.413412e-03\n",
      "    NTN4    -0.097179 3.482262e-05       6.861243e-03\n",
      "    KLK3     0.154404 3.443014e-05       6.861243e-03\n",
      "   AGXT2    -0.132144 3.548818e-05       6.913814e-03\n",
      "  IL17RB    -0.117624 3.703757e-05       7.135493e-03\n",
      "   HNF4A    -0.101051 3.871215e-05       7.376154e-03\n",
      "  SCNN1G    -0.119608 4.013212e-05       7.563595e-03\n",
      "   ITIH2    -0.100441 4.162148e-05       7.759944e-03\n",
      "  SLC2A9    -0.090546 4.311383e-05       7.952666e-03\n",
      "  HS3ST6    -0.111095 4.436392e-05       8.097116e-03\n",
      "  PRIMA1    -0.123677 4.865930e-05       8.782586e-03\n",
      "    SOST    -0.110688 4.913264e-05       8.782586e-03\n",
      " C1QTNF3    -0.085788 5.164640e-05       9.137724e-03\n",
      "    ACY3    -0.140842 5.561366e-05       9.740255e-03\n",
      "   CNNM1    -0.123456 5.772038e-05       9.881156e-03\n",
      " CCDC158    -0.099433 5.812780e-05       9.881156e-03\n",
      "   KCNJ4    -0.080090 5.779124e-05       9.881156e-03\n",
      "    AQP3    -0.096955 6.497647e-05       1.093813e-02\n",
      "   CAPN6    -0.100021 6.620662e-05       1.103804e-02\n",
      "   SPNS2    -0.151406 6.853003e-05       1.131659e-02\n",
      "    ROS1    -0.101173 6.975531e-05       1.141026e-02\n",
      "   PRAP1    -0.116212 7.413112e-05       1.201271e-02\n",
      "   ENPP6    -0.118126 7.614731e-05       1.222517e-02\n",
      "  NECAB3     0.117021 7.730735e-05       1.229754e-02\n",
      "    WNK4    -0.128904 8.499321e-05       1.339725e-02\n",
      "     DDC    -0.124374 8.966106e-05       1.399135e-02\n",
      "  TMEM82    -0.120920 9.037613e-05       1.399135e-02\n",
      "  TM4SF4    -0.073388 9.293262e-05       1.413473e-02\n",
      " SLC39A5    -0.113260 9.270450e-05       1.413473e-02\n",
      "   FOXC1    -0.114300 9.668024e-05       1.457686e-02\n",
      "    ELF5    -0.111808 1.075217e-04       1.607171e-02\n",
      " SLC17A3    -0.109025 1.139158e-04       1.659820e-02\n",
      "  MTNR1A    -0.097369 1.138789e-04       1.659820e-02\n",
      "   ABCB1    -0.119773 1.122985e-04       1.659820e-02\n",
      "   WDR93    -0.087625 1.164673e-04       1.682855e-02\n",
      "  CHRNA4    -0.109294 1.227229e-04       1.758589e-02\n",
      "RAP1GDS1     0.088443 1.264043e-04       1.796495e-02\n",
      "    PBX1    -0.085583 1.324330e-04       1.851820e-02\n",
      "   NPHS2    -0.107072 1.322517e-04       1.851820e-02\n",
      "     DAO    -0.113894 1.349530e-04       1.871960e-02\n",
      "  TUBA4B    -0.087197 1.382105e-04       1.901930e-02\n",
      "  CYP2C9    -0.076400 1.393335e-04       1.902286e-02\n",
      "   MOV10    -0.091924 1.529529e-04       2.071915e-02\n",
      "    TPM1    -0.092005 1.653146e-04       2.202910e-02\n",
      "    NEFM    -0.092020 1.664347e-04       2.202910e-02\n",
      "    ACE2    -0.109600 1.661323e-04       2.202910e-02\n",
      "  AKR1C3    -0.103315 1.712884e-04       2.249977e-02\n",
      "   RASD2     0.083129 1.742663e-04       2.271883e-02\n",
      "   NAT8B    -0.104738 1.773821e-04       2.274991e-02\n",
      "  IGFBP2    -0.118567 1.761260e-04       2.274991e-02\n",
      "   TDRD1    -0.099805 1.784409e-04       2.274991e-02\n",
      "    CRB2    -0.089205 1.802456e-04       2.281225e-02\n",
      "  HAVCR1    -0.094601 1.818410e-04       2.284739e-02\n",
      "  OR2T10    -0.100259 1.945235e-04       2.392087e-02\n",
      "  KLHL32    -0.120470 1.938486e-04       2.392087e-02\n",
      "  SLC7A9    -0.119367 1.926459e-04       2.392087e-02\n",
      "    NAT8    -0.116001 2.130110e-04       2.570103e-02\n",
      " SLC5A10    -0.122104 2.114302e-04       2.570103e-02\n",
      "   IQCA1    -0.078130 2.134465e-04       2.570103e-02\n",
      "     GP2    -0.135633 2.153623e-04       2.575288e-02\n",
      "    RILP     0.103007 2.170357e-04       2.577522e-02\n",
      " CYP4F12    -0.101732 2.220880e-04       2.593195e-02\n",
      "    NOL4    -0.089977 2.213772e-04       2.593195e-02\n",
      "  TUBAL3    -0.107105 2.243378e-04       2.593195e-02\n",
      "    NPR1    -0.108799 2.242799e-04       2.593195e-02\n",
      "SERPINA1    -0.095186 2.391545e-04       2.746158e-02\n",
      "   SOX30    -0.079815 2.452436e-04       2.790574e-02\n",
      "    IQUB    -0.121113 2.462413e-04       2.790574e-02\n",
      "   CLDN2    -0.100366 2.527629e-04       2.845880e-02\n",
      "  LGALS3     0.136428 2.594859e-04       2.884120e-02\n",
      "   OSCP1    -0.108792 2.582411e-04       2.884120e-02\n",
      "  UGT3A2    -0.076028 2.652289e-04       2.929175e-02\n",
      " RIPPLY1    -0.085601 2.716047e-04       2.963072e-02\n",
      "  SLC5A8    -0.118482 2.717160e-04       2.963072e-02\n",
      "SLC22A12    -0.100584 2.815190e-04       3.047877e-02\n",
      "    UMOD    -0.114481 2.830084e-04       3.047877e-02\n",
      "  KCTD16    -0.116398 2.891953e-04       3.095282e-02\n",
      " SLC10A2    -0.096674 2.966541e-04       3.155635e-02\n",
      "  TSPAN1    -0.078527 3.014190e-04       3.167457e-02\n",
      "    GGA2     0.088349 3.008159e-04       3.167457e-02\n",
      "  FBXO15    -0.111926 3.062082e-04       3.198399e-02\n",
      "    GCLC     0.083747 3.231427e-04       3.339404e-02\n",
      "   EFHC2    -0.126637 3.235596e-04       3.339404e-02\n",
      "   HECW1    -0.107962 3.289286e-04       3.366669e-02\n",
      "    COQ2     0.098459 3.300846e-04       3.366669e-02\n",
      "    ADD3    -0.111981 3.370702e-04       3.412337e-02\n",
      "    TAL2    -0.107695 3.384982e-04       3.412337e-02\n",
      "   CLIC5    -0.097197 3.406583e-04       3.414263e-02\n",
      "  GUCA1C    -0.079107 3.523062e-04       3.510711e-02\n",
      "  CCDC78     0.105161 3.650038e-04       3.616458e-02\n",
      "    HAO2    -0.108348 3.755647e-04       3.699952e-02\n",
      "    HCN2     0.127204 3.781196e-04       3.704077e-02\n",
      " ALDH8A1    -0.100357 3.862565e-04       3.762529e-02\n",
      "  TM4SF5    -0.111012 3.914791e-04       3.768557e-02\n",
      "   NPHS1    -0.081825 3.933957e-04       3.768557e-02\n",
      "     HRG    -0.138572 3.911717e-04       3.768557e-02\n",
      "     NTS    -0.096186 3.997367e-04       3.808260e-02\n",
      " SLC13A2    -0.120519 4.140642e-04       3.917125e-02\n",
      "   EPHA7    -0.113431 4.156820e-04       3.917125e-02\n",
      "   LRRC7    -0.087046 4.182720e-04       3.920226e-02\n",
      "  SPOCK2    -0.081693 4.389904e-04       4.018684e-02\n",
      "    CDH9    -0.099420 4.352099e-04       4.018684e-02\n",
      " SLITRK5    -0.087883 4.403656e-04       4.018684e-02\n",
      "    PKLR    -0.107234 4.400375e-04       4.018684e-02\n",
      "  LHFPL3    -0.094429 4.369712e-04       4.018684e-02\n",
      " SLC17A1    -0.104297 4.482433e-04       4.069158e-02\n",
      "    AQP1    -0.096017 4.524170e-04       4.085655e-02\n",
      "    AMPH    -0.102192 4.807037e-04       4.318612e-02\n",
      " SIGLEC8     0.084115 4.872847e-04       4.355170e-02\n",
      "  ATP12A    -0.094677 5.080309e-04       4.517307e-02\n",
      "ARHGAP29    -0.126455 5.168999e-04       4.572718e-02\n",
      "    RBP5    -0.096941 5.367815e-04       4.724494e-02\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/hx/qdnjfcxj2kbfy5k46ywrbrz40000gn/T/ipykernel_4203/2152198575.py:51: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"
     ]
    }
   ],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
