{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "# Set your preferred name\n",
    "USER = \"Jiayi\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output'\n",
    "TRAIT = 'Poor Metabolizer of Drugs'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.928143012Z",
     "start_time": "2024-01-10T21:37:25.918685567Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.554597756Z",
     "start_time": "2024-01-10T21:37:26.530982646Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:26.999521704Z",
     "start_time": "2024-01-10T21:37:26.987471072Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Adrenocortical_Cancer_(ACC)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_ACC_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.ACC.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.180581225Z",
     "start_time": "2024-01-10T21:37:27.217363685Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_rows_and_columns(dataframe, display=False):\n",
    "    \"\"\"\n",
    "    Get the lists of row names and column names of a dataset, and optionally observe them.\n",
    "    :param dataframe:\n",
    "    :param display:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    dataframe_rows = dataframe.index.tolist()\n",
    "    if display:\n",
    "        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n",
    "    dataframe_cols = dataframe.columns.tolist()\n",
    "    if display:\n",
    "        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n",
    "    return dataframe_rows, dataframe_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.185775284Z",
     "start_time": "2024-01-10T21:37:30.181745846Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.269854441Z",
     "start_time": "2024-01-10T21:37:30.184225525Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270036796Z",
     "start_time": "2024-01-10T21:37:30.227180075Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = [ 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = [ 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270232386Z",
     "start_time": "2024-01-10T21:37:30.227282564Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270339062Z",
     "start_time": "2024-01-10T21:37:30.227412883Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270520195Z",
     "start_time": "2024-01-10T21:37:30.227467704Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270626721Z",
     "start_time": "2024-01-10T21:37:30.227530279Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n",
    "    feature_list = []\n",
    "    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_col:\n",
    "        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n",
    "        feature_list.append(age_data)\n",
    "    if gender_col:\n",
    "        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n",
    "        feature_list.append(gender_data)\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=1)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_trait(row_index: str):\n",
    "    \"\"\"\n",
    "    Convert the trait information from Sample IDs to labels depending on the last two digits.\n",
    "    Tumor types range from 01 - 09, normal types from 10 - 19.\n",
    "    :param row_index: the index value of a row\n",
    "    :return: the converted value\n",
    "    \"\"\"\n",
    "    last_two_digits = int(row_index[-2:])\n",
    "\n",
    "    if 1 <= last_two_digits <= 9:\n",
    "        return 1\n",
    "    elif 10 <= last_two_digits <= 19:\n",
    "        return 0\n",
    "    else:\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_age(cell: str):\n",
    "    \"\"\"Convert the cell content about age to a numerical value using regular expression\n",
    "    \"\"\"\n",
    "    match = re.search(r'\\d+', str(cell))\n",
    "    if match:\n",
    "        return int(match.group())\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xena_convert_gender(cell: str):\n",
    "    \"\"\"Convert the cell content about gender to a binary value\n",
    "    \"\"\"\n",
    "    if isinstance(cell, str):\n",
    "        cell = cell.lower()\n",
    "\n",
    "    if cell == \"female\":\n",
    "        return 0\n",
    "    elif cell == \"male\":\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:30.270719762Z",
     "start_time": "2024-01-10T21:37:30.227585330Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.001977877Z",
     "start_time": "2024-01-10T21:37:30.227635342Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import mygene\n",
    "\n",
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.108889749Z",
     "start_time": "2024-01-10T21:38:18.021199718Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:18.112793384Z",
     "start_time": "2024-01-10T21:38:18.107415526Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.377898533Z",
     "start_time": "2024-01-10T21:38:18.111417452Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425336629Z",
     "start_time": "2024-01-10T21:38:31.419043732Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE155985', 'GSE200864', 'GSE205964', 'GSE225292', 'GSE248830']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Poor-Metabolizer-of-Drugs\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:02.935079426Z",
     "start_time": "2024-01-10T21:39:02.923698385Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE248830\\\\GSE248830_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE248830\\\\GSE248830_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious traits\n",
    "cohort = accession_num = \"GSE248830\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE225292\\\\GSE225292_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE225292\\\\GSE225292_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious traits\n",
    "cohort = accession_num = \"GSE225292\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE205964\\\\GSE205964_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE205964\\\\GSE205964_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE205964\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE200864\\\\GSE200864_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE200864\\\\GSE200864_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious traits\n",
    "cohort = accession_num = \"GSE200864\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE155985\\\\GSE155985_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Poor-Metabolizer-of-Drugs\\\\GSE155985\\\\GSE155985_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Stop: No gene mapping\n",
    "cohort = accession_num = \"GSE155985\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepossessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Unlocking Molecular mechanisms and identifying druggable targets in matched-paired brain metastasis of Breast and Lung cancers \"\n",
      "!Series_summary\t\"Introduction: The incidence of brain metastases in cancer patients is increasing, with lung and breast cancer being the most common sources. Despite advancements in targeted therapies, the prognosis remains poor, highlighting the importance to investigate the underlying mechanisms in brain metastases. The aim of this study was to investigate the differences in the molecular mechanisms involved in brain metastasis of breast and lung cancers. In addition, we aimed to identify cancer lineage-specific druggable targets in the brain metastasis. Methods: To that aim, a cohort of 44 FFPE tissue samples, including 22 breast cancer and 22 lung adenocarcinoma (LUAD) and their matched-paired brain metastases were collected. Targeted gene expression profiles of primary tumors were compared to their matched-paired brain metastases samples using nCounter PanCancer IO 360™ Panel of NanoString technologies. Pathway analysis was performed using gene set analysis (GSA) and gene set enrichment analysis (GSEA). The validation was performed by using Immunohistochemistry (IHC) to confirm the expression of immune checkpoint inhibitors. Results: Our results revealed the significant upregulation of cancer-related genes in primary tumors compared to their matched-paired brain metastases (adj. p ≤ 0.05). We found that upregulated differentially expressed genes in breast cancer brain metastasis (BM-BC) and brain metastasis from lung adenocarcinoma (BM-LUAD) were associated with the metabolic stress pathway, particularly related to the glycolysis. Additionally, we found that the upregulated genes in BM-BC and BM-LUAD played roles in immune response regulation, tumor growth, and proliferation. Importantly, we identified high expression of the immune checkpoint VTCN1 in BM-BC, and VISTA, IDO1, NT5E, and HDAC3 in BM-LUAD. Validation using immunohistochemistry further supported these findings. Conclusion: In conclusion, the findings highlight the significance of using matched-paired samples to identify cancer lineage-specific therapies that may improve brain metastasis patients outcomes.\"\n",
      "!Series_overall_design\t\"RNA was extracted from FFPE samples of (primary LUAD and their matched paired brain metastasis n=22, primary BC and their matched paired brain metastasis n=22)\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM7920782</th>\n",
       "      <th>GSM7920783</th>\n",
       "      <th>GSM7920784</th>\n",
       "      <th>GSM7920785</th>\n",
       "      <th>GSM7920786</th>\n",
       "      <th>GSM7920787</th>\n",
       "      <th>GSM7920788</th>\n",
       "      <th>GSM7920789</th>\n",
       "      <th>GSM7920790</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM7920816</th>\n",
       "      <th>GSM7920817</th>\n",
       "      <th>GSM7920818</th>\n",
       "      <th>GSM7920819</th>\n",
       "      <th>GSM7920820</th>\n",
       "      <th>GSM7920821</th>\n",
       "      <th>GSM7920822</th>\n",
       "      <th>GSM7920823</th>\n",
       "      <th>GSM7920824</th>\n",
       "      <th>GSM7920825</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age at diagnosis: 49</td>\n",
       "      <td>age at diagnosis: 44</td>\n",
       "      <td>age at diagnosis: 41</td>\n",
       "      <td>age at diagnosis: 40</td>\n",
       "      <td>age at diagnosis: 48</td>\n",
       "      <td>age at diagnosis: 42</td>\n",
       "      <td>age at diagnosis: 47</td>\n",
       "      <td>age at diagnosis: 53</td>\n",
       "      <td>age at diagnosis: 41</td>\n",
       "      <td>...</td>\n",
       "      <td>age at diagnosis: 63</td>\n",
       "      <td>age at diagnosis: 60</td>\n",
       "      <td>age at diagnosis: 58</td>\n",
       "      <td>age at diagnosis: 70</td>\n",
       "      <td>age at diagnosis: 66</td>\n",
       "      <td>age at diagnosis: 64</td>\n",
       "      <td>age at diagnosis: 60</td>\n",
       "      <td>age at diagnosis: 50</td>\n",
       "      <td>age at diagnosis: 66</td>\n",
       "      <td>age at diagnosis: 74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: female</td>\n",
       "      <td>Sex: male</td>\n",
       "      <td>Sex: male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>histology: TNBC</td>\n",
       "      <td>histology: ER+ PR+ HER2-</td>\n",
       "      <td>histology: Unknown</td>\n",
       "      <td>histology: ER+ PR+ HER2-</td>\n",
       "      <td>histology: TNBC</td>\n",
       "      <td>histology: Unknown</td>\n",
       "      <td>histology: TNBC</td>\n",
       "      <td>histology: ER- PR- HER2+</td>\n",
       "      <td>histology: TNBC</td>\n",
       "      <td>...</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "      <td>histology: adenocaricnoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>smoking status: n.a</td>\n",
       "      <td>...</td>\n",
       "      <td>smoking status: former-smoker</td>\n",
       "      <td>smoking status: former-smoker</td>\n",
       "      <td>smoking status: smoker</td>\n",
       "      <td>smoking status: Never smoking</td>\n",
       "      <td>smoking status: smoker</td>\n",
       "      <td>smoking status: smoker</td>\n",
       "      <td>smoking status: former-smoker</td>\n",
       "      <td>smoking status: unknown</td>\n",
       "      <td>smoking status: smoker</td>\n",
       "      <td>smoking status: former-roker</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>treatment after surgery of bm: surgery + chemo...</td>\n",
       "      <td>treatment after surgery of bm: surgery +  chem...</td>\n",
       "      <td>treatment after surgery of bm: surgery + chemo...</td>\n",
       "      <td>treatment after surgery of bm: surgery</td>\n",
       "      <td>treatment after surgery of bm: surgery</td>\n",
       "      <td>treatment after surgery of bm: surgery +  chem...</td>\n",
       "      <td>treatment after surgery of bm: surgery + chemo...</td>\n",
       "      <td>treatment after surgery of bm: surgery + chemo...</td>\n",
       "      <td>treatment after surgery of bm: surgery + chemo...</td>\n",
       "      <td>...</td>\n",
       "      <td>treatment after surgery of bm: Radiotherapy</td>\n",
       "      <td>treatment after surgery of bm: Radiotherapy</td>\n",
       "      <td>treatment after surgery of bm: Radiotherapy</td>\n",
       "      <td>treatment after surgery of bm: No treatment</td>\n",
       "      <td>treatment after surgery of bm: Radiotherapy</td>\n",
       "      <td>treatment after surgery of bm: WBRT</td>\n",
       "      <td>treatment after surgery of bm: SRT</td>\n",
       "      <td>treatment after surgery of bm: WBRT</td>\n",
       "      <td>treatment after surgery of bm: Radiotherapy</td>\n",
       "      <td>treatment after surgery of bm: Radiotherapy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "\n",
       "                                          GSM7920782  \\\n",
       "0                               age at diagnosis: 49   \n",
       "1                                        Sex: female   \n",
       "2                                    histology: TNBC   \n",
       "3                                smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery + chemo...   \n",
       "\n",
       "                                          GSM7920783  \\\n",
       "0                               age at diagnosis: 44   \n",
       "1                                        Sex: female   \n",
       "2                           histology: ER+ PR+ HER2-   \n",
       "3                                smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery +  chem...   \n",
       "\n",
       "                                          GSM7920784  \\\n",
       "0                               age at diagnosis: 41   \n",
       "1                                        Sex: female   \n",
       "2                                 histology: Unknown   \n",
       "3                                smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery + chemo...   \n",
       "\n",
       "                               GSM7920785  \\\n",
       "0                    age at diagnosis: 40   \n",
       "1                             Sex: female   \n",
       "2                histology: ER+ PR+ HER2-   \n",
       "3                     smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery   \n",
       "\n",
       "                               GSM7920786  \\\n",
       "0                    age at diagnosis: 48   \n",
       "1                             Sex: female   \n",
       "2                         histology: TNBC   \n",
       "3                     smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery   \n",
       "\n",
       "                                          GSM7920787  \\\n",
       "0                               age at diagnosis: 42   \n",
       "1                                        Sex: female   \n",
       "2                                 histology: Unknown   \n",
       "3                                smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery +  chem...   \n",
       "\n",
       "                                          GSM7920788  \\\n",
       "0                               age at diagnosis: 47   \n",
       "1                                        Sex: female   \n",
       "2                                    histology: TNBC   \n",
       "3                                smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery + chemo...   \n",
       "\n",
       "                                          GSM7920789  \\\n",
       "0                               age at diagnosis: 53   \n",
       "1                                        Sex: female   \n",
       "2                           histology: ER- PR- HER2+   \n",
       "3                                smoking status: n.a   \n",
       "4  treatment after surgery of bm: surgery + chemo...   \n",
       "\n",
       "                                          GSM7920790  ...  \\\n",
       "0                               age at diagnosis: 41  ...   \n",
       "1                                        Sex: female  ...   \n",
       "2                                    histology: TNBC  ...   \n",
       "3                                smoking status: n.a  ...   \n",
       "4  treatment after surgery of bm: surgery + chemo...  ...   \n",
       "\n",
       "                                    GSM7920816  \\\n",
       "0                         age at diagnosis: 63   \n",
       "1                                    Sex: male   \n",
       "2                    histology: adenocaricnoma   \n",
       "3                smoking status: former-smoker   \n",
       "4  treatment after surgery of bm: Radiotherapy   \n",
       "\n",
       "                                    GSM7920817  \\\n",
       "0                         age at diagnosis: 60   \n",
       "1                                    Sex: male   \n",
       "2                    histology: adenocaricnoma   \n",
       "3                smoking status: former-smoker   \n",
       "4  treatment after surgery of bm: Radiotherapy   \n",
       "\n",
       "                                    GSM7920818  \\\n",
       "0                         age at diagnosis: 58   \n",
       "1                                  Sex: female   \n",
       "2                    histology: adenocaricnoma   \n",
       "3                       smoking status: smoker   \n",
       "4  treatment after surgery of bm: Radiotherapy   \n",
       "\n",
       "                                    GSM7920819  \\\n",
       "0                         age at diagnosis: 70   \n",
       "1                                  Sex: female   \n",
       "2                    histology: adenocaricnoma   \n",
       "3                smoking status: Never smoking   \n",
       "4  treatment after surgery of bm: No treatment   \n",
       "\n",
       "                                    GSM7920820  \\\n",
       "0                         age at diagnosis: 66   \n",
       "1                                  Sex: female   \n",
       "2                    histology: adenocaricnoma   \n",
       "3                       smoking status: smoker   \n",
       "4  treatment after surgery of bm: Radiotherapy   \n",
       "\n",
       "                            GSM7920821                          GSM7920822  \\\n",
       "0                 age at diagnosis: 64                age at diagnosis: 60   \n",
       "1                            Sex: male                           Sex: male   \n",
       "2            histology: adenocaricnoma           histology: adenocaricnoma   \n",
       "3               smoking status: smoker       smoking status: former-smoker   \n",
       "4  treatment after surgery of bm: WBRT  treatment after surgery of bm: SRT   \n",
       "\n",
       "                            GSM7920823  \\\n",
       "0                 age at diagnosis: 50   \n",
       "1                          Sex: female   \n",
       "2            histology: adenocaricnoma   \n",
       "3              smoking status: unknown   \n",
       "4  treatment after surgery of bm: WBRT   \n",
       "\n",
       "                                    GSM7920824  \\\n",
       "0                         age at diagnosis: 66   \n",
       "1                                    Sex: male   \n",
       "2                    histology: adenocaricnoma   \n",
       "3                       smoking status: smoker   \n",
       "4  treatment after surgery of bm: Radiotherapy   \n",
       "\n",
       "                                    GSM7920825  \n",
       "0                         age at diagnosis: 74  \n",
       "1                                    Sex: male  \n",
       "2                    histology: adenocaricnoma  \n",
       "3                 smoking status: former-roker  \n",
       "4  treatment after surgery of bm: Radiotherapy  \n",
       "\n",
       "[5 rows x 45 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from utils import *\n",
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)\n",
    "\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'histology: TNBC',\n",
       "       'histology: ER+ PR+ HER2-', 'histology: Unknown',\n",
       "       'histology: ER- PR- HER2+', 'histology: ER+ PR-HER2+',\n",
       "       'histology: ER+ PR- HER2-', 'histology: ER- PR+ HER2-',\n",
       "       'histology: adenocaricnoma'], dtype=object)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "trait_type = 'binary'\n",
    "\n",
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "# 这个函数将组织类型（tissue type）转换为有关癫痫存在与否的二进制值。\n",
    "# 它是基于特定的假设，即如果组织类型是“胰腺导管腺癌”（Pancreatic Ductal Adenocarcinoma），则认为癫痫存在（返回1）；否则，认为癫痫不存在（返回0）。\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if 'KIN1148' in tissue_type:\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "# def convert_trait(tumor_grade):\n",
    "#     if (tumor_grade == 'tumor grade: 2' or tumor_grade == 'tumor grade: 3' or tumor_grade == 'tumor grade: 4'):\n",
    "#         return 1  \n",
    "#     elif tumor_grade == 'tumor grade: 1':\n",
    "#         return 0  \n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# 这个函数的目的是将年龄的字符串表示转换为一个连续的数值型表示。如果年龄未知（例如，标记为'n.a.'），则返回None。\n",
    "# 函数尝试从传入的字符串中提取出一个整数作为年龄值。如果字符串的格式不符合预期，导致提取失败，同样返回None。\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(':')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# 这个函数将性别的字符串表示转换为二进制值，其中“female”对应1，“male”对应0。如果性别未知或字符串不符合预期格式，则返回None。\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\Jiayi\\..\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM6236086</th>\n",
       "      <th>GSM6236087</th>\n",
       "      <th>GSM6236088</th>\n",
       "      <th>GSM6236089</th>\n",
       "      <th>GSM6236090</th>\n",
       "      <th>GSM6236091</th>\n",
       "      <th>GSM6236092</th>\n",
       "      <th>GSM6236093</th>\n",
       "      <th>GSM6236094</th>\n",
       "      <th>GSM6236095</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM6236106</th>\n",
       "      <th>GSM6236107</th>\n",
       "      <th>GSM6236108</th>\n",
       "      <th>GSM6236109</th>\n",
       "      <th>GSM6236110</th>\n",
       "      <th>GSM6236111</th>\n",
       "      <th>GSM6236112</th>\n",
       "      <th>GSM6236113</th>\n",
       "      <th>GSM6236114</th>\n",
       "      <th>GSM6236115</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Poor Metabolizer of Drugs</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           GSM6236086  GSM6236087  GSM6236088  GSM6236089  \\\n",
       "Poor Metabolizer of Drugs           0           0           0           0   \n",
       "\n",
       "                           GSM6236090  GSM6236091  GSM6236092  GSM6236093  \\\n",
       "Poor Metabolizer of Drugs           1           1           0           0   \n",
       "\n",
       "                           GSM6236094  GSM6236095  ...  GSM6236106  \\\n",
       "Poor Metabolizer of Drugs           0           0  ...           1   \n",
       "\n",
       "                           GSM6236107  GSM6236108  GSM6236109  GSM6236110  \\\n",
       "Poor Metabolizer of Drugs           1           1           0           0   \n",
       "\n",
       "                           GSM6236111  GSM6236112  GSM6236113  GSM6236114  \\\n",
       "Poor Metabolizer of Drugs           0           0           0           0   \n",
       "\n",
       "                           GSM6236115  \n",
       "Poor Metabolizer of Drugs           0  \n",
       "\n",
       "[1 rows x 30 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM6236086</th>\n",
       "      <th>GSM6236087</th>\n",
       "      <th>GSM6236088</th>\n",
       "      <th>GSM6236089</th>\n",
       "      <th>GSM6236090</th>\n",
       "      <th>GSM6236091</th>\n",
       "      <th>GSM6236092</th>\n",
       "      <th>GSM6236093</th>\n",
       "      <th>GSM6236094</th>\n",
       "      <th>GSM6236095</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM6236106</th>\n",
       "      <th>GSM6236107</th>\n",
       "      <th>GSM6236108</th>\n",
       "      <th>GSM6236109</th>\n",
       "      <th>GSM6236110</th>\n",
       "      <th>GSM6236111</th>\n",
       "      <th>GSM6236112</th>\n",
       "      <th>GSM6236113</th>\n",
       "      <th>GSM6236114</th>\n",
       "      <th>GSM6236115</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_19_P00315452</th>\n",
       "      <td>7.173760</td>\n",
       "      <td>6.762991</td>\n",
       "      <td>6.914186</td>\n",
       "      <td>6.937423</td>\n",
       "      <td>6.792031</td>\n",
       "      <td>6.898450</td>\n",
       "      <td>6.639883</td>\n",
       "      <td>6.598177</td>\n",
       "      <td>5.969626</td>\n",
       "      <td>6.888337</td>\n",
       "      <td>...</td>\n",
       "      <td>6.493188</td>\n",
       "      <td>6.714703</td>\n",
       "      <td>7.234498</td>\n",
       "      <td>6.505150</td>\n",
       "      <td>6.451211</td>\n",
       "      <td>6.414192</td>\n",
       "      <td>6.329722</td>\n",
       "      <td>5.881012</td>\n",
       "      <td>5.950313</td>\n",
       "      <td>5.749534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_19_P00315459</th>\n",
       "      <td>6.641811</td>\n",
       "      <td>6.941341</td>\n",
       "      <td>6.773469</td>\n",
       "      <td>6.826125</td>\n",
       "      <td>6.889960</td>\n",
       "      <td>6.820392</td>\n",
       "      <td>7.176838</td>\n",
       "      <td>6.832257</td>\n",
       "      <td>6.948757</td>\n",
       "      <td>6.717219</td>\n",
       "      <td>...</td>\n",
       "      <td>7.288020</td>\n",
       "      <td>6.855076</td>\n",
       "      <td>6.878562</td>\n",
       "      <td>6.905086</td>\n",
       "      <td>7.413628</td>\n",
       "      <td>7.300429</td>\n",
       "      <td>7.250140</td>\n",
       "      <td>7.134940</td>\n",
       "      <td>7.112613</td>\n",
       "      <td>7.188424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_19_P00315493</th>\n",
       "      <td>5.867896</td>\n",
       "      <td>6.135453</td>\n",
       "      <td>5.789642</td>\n",
       "      <td>5.962318</td>\n",
       "      <td>6.013834</td>\n",
       "      <td>6.278759</td>\n",
       "      <td>6.374460</td>\n",
       "      <td>6.337176</td>\n",
       "      <td>6.166247</td>\n",
       "      <td>6.008242</td>\n",
       "      <td>...</td>\n",
       "      <td>6.409391</td>\n",
       "      <td>5.977662</td>\n",
       "      <td>6.170259</td>\n",
       "      <td>5.911292</td>\n",
       "      <td>6.353147</td>\n",
       "      <td>6.229780</td>\n",
       "      <td>6.437128</td>\n",
       "      <td>7.020239</td>\n",
       "      <td>6.625709</td>\n",
       "      <td>6.992466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_19_P00315506</th>\n",
       "      <td>8.294085</td>\n",
       "      <td>7.990482</td>\n",
       "      <td>8.138528</td>\n",
       "      <td>8.307504</td>\n",
       "      <td>8.024863</td>\n",
       "      <td>8.454231</td>\n",
       "      <td>7.589339</td>\n",
       "      <td>8.801331</td>\n",
       "      <td>10.506275</td>\n",
       "      <td>8.022183</td>\n",
       "      <td>...</td>\n",
       "      <td>8.347031</td>\n",
       "      <td>8.015554</td>\n",
       "      <td>8.214076</td>\n",
       "      <td>9.004618</td>\n",
       "      <td>7.971448</td>\n",
       "      <td>7.831149</td>\n",
       "      <td>7.744722</td>\n",
       "      <td>11.255953</td>\n",
       "      <td>11.249874</td>\n",
       "      <td>11.102544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_19_P00315550</th>\n",
       "      <td>6.956328</td>\n",
       "      <td>7.036906</td>\n",
       "      <td>7.139551</td>\n",
       "      <td>7.117730</td>\n",
       "      <td>7.005625</td>\n",
       "      <td>7.126704</td>\n",
       "      <td>6.778515</td>\n",
       "      <td>7.036174</td>\n",
       "      <td>7.313488</td>\n",
       "      <td>6.817730</td>\n",
       "      <td>...</td>\n",
       "      <td>6.849040</td>\n",
       "      <td>6.693487</td>\n",
       "      <td>6.740905</td>\n",
       "      <td>6.788120</td>\n",
       "      <td>7.292168</td>\n",
       "      <td>6.959809</td>\n",
       "      <td>7.117383</td>\n",
       "      <td>7.283243</td>\n",
       "      <td>7.091524</td>\n",
       "      <td>7.178665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_33_P3888365</th>\n",
       "      <td>8.944663</td>\n",
       "      <td>8.850578</td>\n",
       "      <td>9.032344</td>\n",
       "      <td>9.044030</td>\n",
       "      <td>8.848727</td>\n",
       "      <td>8.844967</td>\n",
       "      <td>8.881420</td>\n",
       "      <td>8.999037</td>\n",
       "      <td>9.218624</td>\n",
       "      <td>8.987430</td>\n",
       "      <td>...</td>\n",
       "      <td>8.757612</td>\n",
       "      <td>8.773469</td>\n",
       "      <td>8.795336</td>\n",
       "      <td>8.907216</td>\n",
       "      <td>8.908743</td>\n",
       "      <td>8.900011</td>\n",
       "      <td>8.907942</td>\n",
       "      <td>8.864393</td>\n",
       "      <td>8.905086</td>\n",
       "      <td>8.866403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_33_P3888485</th>\n",
       "      <td>6.152454</td>\n",
       "      <td>6.090818</td>\n",
       "      <td>5.953808</td>\n",
       "      <td>6.139210</td>\n",
       "      <td>6.013834</td>\n",
       "      <td>6.014206</td>\n",
       "      <td>6.104337</td>\n",
       "      <td>6.035624</td>\n",
       "      <td>6.300429</td>\n",
       "      <td>6.224002</td>\n",
       "      <td>...</td>\n",
       "      <td>6.123777</td>\n",
       "      <td>7.133228</td>\n",
       "      <td>5.901671</td>\n",
       "      <td>6.211564</td>\n",
       "      <td>5.930343</td>\n",
       "      <td>6.372140</td>\n",
       "      <td>5.965784</td>\n",
       "      <td>6.004127</td>\n",
       "      <td>6.166247</td>\n",
       "      <td>6.151947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_33_P3889862</th>\n",
       "      <td>6.186857</td>\n",
       "      <td>6.279533</td>\n",
       "      <td>6.395463</td>\n",
       "      <td>6.398601</td>\n",
       "      <td>5.965784</td>\n",
       "      <td>6.412499</td>\n",
       "      <td>5.922436</td>\n",
       "      <td>6.337176</td>\n",
       "      <td>6.166247</td>\n",
       "      <td>6.166916</td>\n",
       "      <td>...</td>\n",
       "      <td>6.029747</td>\n",
       "      <td>6.258330</td>\n",
       "      <td>6.231701</td>\n",
       "      <td>6.262721</td>\n",
       "      <td>6.412217</td>\n",
       "      <td>6.114436</td>\n",
       "      <td>6.154649</td>\n",
       "      <td>6.232980</td>\n",
       "      <td>6.437960</td>\n",
       "      <td>6.343408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_33_P3893191</th>\n",
       "      <td>6.986411</td>\n",
       "      <td>7.561606</td>\n",
       "      <td>6.878971</td>\n",
       "      <td>7.239360</td>\n",
       "      <td>7.074498</td>\n",
       "      <td>7.355792</td>\n",
       "      <td>7.768295</td>\n",
       "      <td>7.546380</td>\n",
       "      <td>8.117080</td>\n",
       "      <td>7.007121</td>\n",
       "      <td>...</td>\n",
       "      <td>7.430034</td>\n",
       "      <td>7.180406</td>\n",
       "      <td>7.202124</td>\n",
       "      <td>7.157178</td>\n",
       "      <td>8.071775</td>\n",
       "      <td>8.231141</td>\n",
       "      <td>8.036036</td>\n",
       "      <td>8.601616</td>\n",
       "      <td>7.826760</td>\n",
       "      <td>8.306138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_33_P3897698</th>\n",
       "      <td>6.705056</td>\n",
       "      <td>6.890366</td>\n",
       "      <td>6.507002</td>\n",
       "      <td>6.851020</td>\n",
       "      <td>6.757557</td>\n",
       "      <td>7.016437</td>\n",
       "      <td>6.639883</td>\n",
       "      <td>6.935263</td>\n",
       "      <td>6.662728</td>\n",
       "      <td>6.717219</td>\n",
       "      <td>...</td>\n",
       "      <td>6.696968</td>\n",
       "      <td>6.615936</td>\n",
       "      <td>6.759555</td>\n",
       "      <td>6.640004</td>\n",
       "      <td>6.877949</td>\n",
       "      <td>6.709314</td>\n",
       "      <td>6.818689</td>\n",
       "      <td>6.668885</td>\n",
       "      <td>6.241268</td>\n",
       "      <td>6.524346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23478 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GSM6236086  GSM6236087  GSM6236088  GSM6236089  GSM6236090  \\\n",
       "ID                                                                           \n",
       "A_19_P00315452    7.173760    6.762991    6.914186    6.937423    6.792031   \n",
       "A_19_P00315459    6.641811    6.941341    6.773469    6.826125    6.889960   \n",
       "A_19_P00315493    5.867896    6.135453    5.789642    5.962318    6.013834   \n",
       "A_19_P00315506    8.294085    7.990482    8.138528    8.307504    8.024863   \n",
       "A_19_P00315550    6.956328    7.036906    7.139551    7.117730    7.005625   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "A_33_P3888365     8.944663    8.850578    9.032344    9.044030    8.848727   \n",
       "A_33_P3888485     6.152454    6.090818    5.953808    6.139210    6.013834   \n",
       "A_33_P3889862     6.186857    6.279533    6.395463    6.398601    5.965784   \n",
       "A_33_P3893191     6.986411    7.561606    6.878971    7.239360    7.074498   \n",
       "A_33_P3897698     6.705056    6.890366    6.507002    6.851020    6.757557   \n",
       "\n",
       "                GSM6236091  GSM6236092  GSM6236093  GSM6236094  GSM6236095  \\\n",
       "ID                                                                           \n",
       "A_19_P00315452    6.898450    6.639883    6.598177    5.969626    6.888337   \n",
       "A_19_P00315459    6.820392    7.176838    6.832257    6.948757    6.717219   \n",
       "A_19_P00315493    6.278759    6.374460    6.337176    6.166247    6.008242   \n",
       "A_19_P00315506    8.454231    7.589339    8.801331   10.506275    8.022183   \n",
       "A_19_P00315550    7.126704    6.778515    7.036174    7.313488    6.817730   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "A_33_P3888365     8.844967    8.881420    8.999037    9.218624    8.987430   \n",
       "A_33_P3888485     6.014206    6.104337    6.035624    6.300429    6.224002   \n",
       "A_33_P3889862     6.412499    5.922436    6.337176    6.166247    6.166916   \n",
       "A_33_P3893191     7.355792    7.768295    7.546380    8.117080    7.007121   \n",
       "A_33_P3897698     7.016437    6.639883    6.935263    6.662728    6.717219   \n",
       "\n",
       "                ...  GSM6236106  GSM6236107  GSM6236108  GSM6236109  \\\n",
       "ID              ...                                                   \n",
       "A_19_P00315452  ...    6.493188    6.714703    7.234498    6.505150   \n",
       "A_19_P00315459  ...    7.288020    6.855076    6.878562    6.905086   \n",
       "A_19_P00315493  ...    6.409391    5.977662    6.170259    5.911292   \n",
       "A_19_P00315506  ...    8.347031    8.015554    8.214076    9.004618   \n",
       "A_19_P00315550  ...    6.849040    6.693487    6.740905    6.788120   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "A_33_P3888365   ...    8.757612    8.773469    8.795336    8.907216   \n",
       "A_33_P3888485   ...    6.123777    7.133228    5.901671    6.211564   \n",
       "A_33_P3889862   ...    6.029747    6.258330    6.231701    6.262721   \n",
       "A_33_P3893191   ...    7.430034    7.180406    7.202124    7.157178   \n",
       "A_33_P3897698   ...    6.696968    6.615936    6.759555    6.640004   \n",
       "\n",
       "                GSM6236110  GSM6236111  GSM6236112  GSM6236113  GSM6236114  \\\n",
       "ID                                                                           \n",
       "A_19_P00315452    6.451211    6.414192    6.329722    5.881012    5.950313   \n",
       "A_19_P00315459    7.413628    7.300429    7.250140    7.134940    7.112613   \n",
       "A_19_P00315493    6.353147    6.229780    6.437128    7.020239    6.625709   \n",
       "A_19_P00315506    7.971448    7.831149    7.744722   11.255953   11.249874   \n",
       "A_19_P00315550    7.292168    6.959809    7.117383    7.283243    7.091524   \n",
       "...                    ...         ...         ...         ...         ...   \n",
       "A_33_P3888365     8.908743    8.900011    8.907942    8.864393    8.905086   \n",
       "A_33_P3888485     5.930343    6.372140    5.965784    6.004127    6.166247   \n",
       "A_33_P3889862     6.412217    6.114436    6.154649    6.232980    6.437960   \n",
       "A_33_P3893191     8.071775    8.231141    8.036036    8.601616    7.826760   \n",
       "A_33_P3897698     6.877949    6.709314    6.818689    6.668885    6.241268   \n",
       "\n",
       "                GSM6236115  \n",
       "ID                          \n",
       "A_19_P00315452    5.749534  \n",
       "A_19_P00315459    7.188424  \n",
       "A_19_P00315493    6.992466  \n",
       "A_19_P00315506   11.102544  \n",
       "A_19_P00315550    7.178665  \n",
       "...                    ...  \n",
       "A_33_P3888365     8.866403  \n",
       "A_33_P3888485     6.151947  \n",
       "A_33_P3889862     6.343408  \n",
       "A_33_P3893191     8.306138  \n",
       "A_33_P3897698     6.524346  \n",
       "\n",
       "[23478 rows x 30 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['GE_BrightCorner', 'DarkCorner', 'A_23_P117082', 'A_33_P3246448', 'A_33_P3318220'], 'SPOT_ID': ['CONTROL', 'CONTROL', 'A_23_P117082', 'A_33_P3246448', 'A_33_P3318220'], 'CONTROL_TYPE': ['pos', 'pos', 'FALSE', 'FALSE', 'FALSE'], 'REFSEQ': [nan, nan, 'NM_015987', 'NM_080671', 'NM_178466'], 'GB_ACC': [nan, nan, 'NM_015987', 'NM_080671', 'NM_178466'], 'LOCUSLINK_ID': [nan, nan, 50865.0, 23704.0, 128861.0], 'GENE_SYMBOL': [nan, nan, 'HEBP1', 'KCNE4', 'BPIFA3'], 'GENE_NAME': [nan, nan, 'heme binding protein 1', 'potassium voltage-gated channel, Isk-related family, member 4', 'BPI fold containing family A, member 3'], 'UNIGENE_ID': [nan, nan, 'Hs.642618', 'Hs.348522', 'Hs.360989'], 'ENSEMBL_ID': [nan, nan, 'ENST00000014930', 'ENST00000281830', 'ENST00000375454'], 'ACCESSION_STRING': [nan, nan, 'ref|NM_015987|ens|ENST00000014930|gb|AF117615|gb|BC016277', 'ref|NM_080671|ens|ENST00000281830|tc|THC2655788', 'ref|NM_178466|ens|ENST00000375454|ens|ENST00000471233|tc|THC2478474'], 'CHROMOSOMAL_LOCATION': [nan, nan, 'chr12:13127906-13127847', 'chr2:223920197-223920256', 'chr20:31812208-31812267'], 'CYTOBAND': [nan, nan, 'hs|12p13.1', 'hs|2q36.1', 'hs|20q11.21'], 'DESCRIPTION': [nan, nan, 'Homo sapiens heme binding protein 1 (HEBP1), mRNA [NM_015987]', 'Homo sapiens potassium voltage-gated channel, Isk-related family, member 4 (KCNE4), mRNA [NM_080671]', 'Homo sapiens BPI fold containing family A, member 3 (BPIFA3), transcript variant 1, mRNA [NM_178466]'], 'GO_ID': [nan, nan, 'GO:0005488(binding)|GO:0005576(extracellular region)|GO:0005737(cytoplasm)|GO:0005739(mitochondrion)|GO:0005829(cytosol)|GO:0007623(circadian rhythm)|GO:0020037(heme binding)', 'GO:0005244(voltage-gated ion channel activity)|GO:0005249(voltage-gated potassium channel activity)|GO:0006811(ion transport)|GO:0006813(potassium ion transport)|GO:0016020(membrane)|GO:0016021(integral to membrane)|GO:0016324(apical plasma membrane)', 'GO:0005576(extracellular region)|GO:0008289(lipid binding)'], 'SEQUENCE': [nan, nan, 'AAGGGGGAAAATGTGATTTGTGCCTGATCTTTCATCTGTGATTCTTATAAGAGCTTTGTC', 'GCAAGTCTCTCTGCACCTATTAAAAAGTGATGTATATACTTCCTTCTTATTCTGTTGAGT', 'CATTCCATAAGGAGTGGTTCTCGGCAAATATCTCACTTGAATTTGACCTTGAATTGAGAC']}\n",
      "\n",
      "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
      "    ['A_19_P00315452', 'A_19_P00315459', 'A_19_P00315493', 'A_19_P00315506', 'A_19_P00315550', 'A_19_P00315551', 'A_19_P00315581', 'A_19_P00315583', 'A_19_P00315584', 'A_19_P00315601', 'A_19_P00315603', 'A_19_P00315649', 'A_19_P00315668', 'A_19_P00315693', 'A_19_P00315705', 'A_19_P00315716', 'A_19_P00315717', 'A_19_P00315718', 'A_19_P00315738', 'A_19_P00315753']\n",
      "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
      "    identifier_key = 'key_name1'\n",
      "    gene_symbol_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['GE_BrightCorner', 'DarkCorner', 'A_23_P117082', 'A_33_P3246448', 'A_33_P3318220'], 'SPOT_ID': ['CONTROL', 'CONTROL', 'A_23_P117082', 'A_33_P3246448', 'A_33_P3318220'], 'CONTROL_TYPE': ['pos', 'pos', 'FALSE', 'FALSE', 'FALSE'], 'REFSEQ': [nan, nan, 'NM_015987', 'NM_080671', 'NM_178466'], 'GB_ACC': [nan, nan, 'NM_015987', 'NM_080671', 'NM_178466'], 'LOCUSLINK_ID': [nan, nan, 50865.0, 23704.0, 128861.0], 'GENE_SYMBOL': [nan, nan, 'HEBP1', 'KCNE4', 'BPIFA3'], 'GENE_NAME': [nan, nan, 'heme binding protein 1', 'potassium voltage-gated channel, Isk-related family, member 4', 'BPI fold containing family A, member 3'], 'UNIGENE_ID': [nan, nan, 'Hs.642618', 'Hs.348522', 'Hs.360989'], 'ENSEMBL_ID': [nan, nan, 'ENST00000014930', 'ENST00000281830', 'ENST00000375454'], 'ACCESSION_STRING': [nan, nan, 'ref|NM_015987|ens|ENST00000014930|gb|AF117615|gb|BC016277', 'ref|NM_080671|ens|ENST00000281830|tc|THC2655788', 'ref|NM_178466|ens|ENST00000375454|ens|ENST00000471233|tc|THC2478474'], 'CHROMOSOMAL_LOCATION': [nan, nan, 'chr12:13127906-13127847', 'chr2:223920197-223920256', 'chr20:31812208-31812267'], 'CYTOBAND': [nan, nan, 'hs|12p13.1', 'hs|2q36.1', 'hs|20q11.21'], 'DESCRIPTION': [nan, nan, 'Homo sapiens heme binding protein 1 (HEBP1), mRNA [NM_015987]', 'Homo sapiens potassium voltage-gated channel, Isk-related family, member 4 (KCNE4), mRNA [NM_080671]', 'Homo sapiens BPI fold containing family A, member 3 (BPIFA3), transcript variant 1, mRNA [NM_178466]'], 'GO_ID': [nan, nan, 'GO:0005488(binding)|GO:0005576(extracellular region)|GO:0005737(cytoplasm)|GO:0005739(mitochondrion)|GO:0005829(cytosol)|GO:0007623(circadian rhythm)|GO:0020037(heme binding)', 'GO:0005244(voltage-gated ion channel activity)|GO:0005249(voltage-gated potassium channel activity)|GO:0006811(ion transport)|GO:0006813(potassium ion transport)|GO:0016020(membrane)|GO:0016021(integral to membrane)|GO:0016324(apical plasma membrane)', 'GO:0005576(extracellular region)|GO:0008289(lipid binding)'], 'SEQUENCE': [nan, nan, 'AAGGGGGAAAATGTGATTTGTGCCTGATCTTTCATCTGTGATTCTTATAAGAGCTTTGTC', 'GCAAGTCTCTCTGCACCTATTAAAAAGTGATGTATATACTTCCTTCTTATTCTGTTGAGT', 'CATTCCATAAGGAGTGGTTCTCGGCAAATATCTCACTTGAATTTGACCTTGAATTGAGAC']}\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'SPOT_ID', 'CONTROL_TYPE', 'REFSEQ', 'GB_ACC', 'LOCUSLINK_ID',\n",
       "       'GENE_SYMBOL', 'GENE_NAME', 'UNIGENE_ID', 'ENSEMBL_ID',\n",
       "       'ACCESSION_STRING', 'CHROMOSOMAL_LOCATION', 'CYTOBAND', 'DESCRIPTION',\n",
       "       'GO_ID', 'SEQUENCE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'GENE_SYMBOL'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM6236086</th>\n",
       "      <th>GSM6236087</th>\n",
       "      <th>GSM6236088</th>\n",
       "      <th>GSM6236089</th>\n",
       "      <th>GSM6236090</th>\n",
       "      <th>GSM6236091</th>\n",
       "      <th>GSM6236092</th>\n",
       "      <th>GSM6236093</th>\n",
       "      <th>GSM6236094</th>\n",
       "      <th>GSM6236095</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM6236106</th>\n",
       "      <th>GSM6236107</th>\n",
       "      <th>GSM6236108</th>\n",
       "      <th>GSM6236109</th>\n",
       "      <th>GSM6236110</th>\n",
       "      <th>GSM6236111</th>\n",
       "      <th>GSM6236112</th>\n",
       "      <th>GSM6236113</th>\n",
       "      <th>GSM6236114</th>\n",
       "      <th>GSM6236115</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>8.396817</td>\n",
       "      <td>8.540113</td>\n",
       "      <td>8.272503</td>\n",
       "      <td>8.261481</td>\n",
       "      <td>8.038747</td>\n",
       "      <td>8.073204</td>\n",
       "      <td>8.005605</td>\n",
       "      <td>8.016628</td>\n",
       "      <td>7.865073</td>\n",
       "      <td>8.250083</td>\n",
       "      <td>...</td>\n",
       "      <td>8.218620</td>\n",
       "      <td>8.115851</td>\n",
       "      <td>8.224858</td>\n",
       "      <td>8.079500</td>\n",
       "      <td>7.907925</td>\n",
       "      <td>8.315434</td>\n",
       "      <td>7.946120</td>\n",
       "      <td>7.287691</td>\n",
       "      <td>7.422345</td>\n",
       "      <td>7.434011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>8.569792</td>\n",
       "      <td>8.566561</td>\n",
       "      <td>8.401626</td>\n",
       "      <td>8.390599</td>\n",
       "      <td>8.284091</td>\n",
       "      <td>8.243293</td>\n",
       "      <td>8.365375</td>\n",
       "      <td>8.403580</td>\n",
       "      <td>8.628567</td>\n",
       "      <td>8.536636</td>\n",
       "      <td>...</td>\n",
       "      <td>8.360774</td>\n",
       "      <td>8.144956</td>\n",
       "      <td>8.126360</td>\n",
       "      <td>8.305530</td>\n",
       "      <td>7.940901</td>\n",
       "      <td>8.089318</td>\n",
       "      <td>8.105210</td>\n",
       "      <td>8.017876</td>\n",
       "      <td>7.921990</td>\n",
       "      <td>7.906390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>15.252783</td>\n",
       "      <td>15.275463</td>\n",
       "      <td>15.530561</td>\n",
       "      <td>15.455341</td>\n",
       "      <td>15.120197</td>\n",
       "      <td>15.357978</td>\n",
       "      <td>14.597179</td>\n",
       "      <td>15.539549</td>\n",
       "      <td>15.079838</td>\n",
       "      <td>15.541395</td>\n",
       "      <td>...</td>\n",
       "      <td>14.793671</td>\n",
       "      <td>15.200568</td>\n",
       "      <td>15.416836</td>\n",
       "      <td>15.520590</td>\n",
       "      <td>14.606018</td>\n",
       "      <td>14.731785</td>\n",
       "      <td>14.648498</td>\n",
       "      <td>14.460178</td>\n",
       "      <td>14.779647</td>\n",
       "      <td>14.665974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>11.209895</td>\n",
       "      <td>11.128317</td>\n",
       "      <td>11.365849</td>\n",
       "      <td>10.987619</td>\n",
       "      <td>11.215685</td>\n",
       "      <td>10.982720</td>\n",
       "      <td>11.058364</td>\n",
       "      <td>11.125532</td>\n",
       "      <td>11.145937</td>\n",
       "      <td>10.993670</td>\n",
       "      <td>...</td>\n",
       "      <td>11.404432</td>\n",
       "      <td>11.664751</td>\n",
       "      <td>11.584156</td>\n",
       "      <td>11.537469</td>\n",
       "      <td>12.564070</td>\n",
       "      <td>12.512469</td>\n",
       "      <td>12.329296</td>\n",
       "      <td>12.028234</td>\n",
       "      <td>12.322689</td>\n",
       "      <td>12.130088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>8.399242</td>\n",
       "      <td>8.117124</td>\n",
       "      <td>8.307808</td>\n",
       "      <td>8.230821</td>\n",
       "      <td>8.149916</td>\n",
       "      <td>8.197217</td>\n",
       "      <td>8.146950</td>\n",
       "      <td>8.126790</td>\n",
       "      <td>7.851124</td>\n",
       "      <td>8.137845</td>\n",
       "      <td>...</td>\n",
       "      <td>8.040290</td>\n",
       "      <td>8.132457</td>\n",
       "      <td>8.359567</td>\n",
       "      <td>7.994825</td>\n",
       "      <td>8.121188</td>\n",
       "      <td>8.262173</td>\n",
       "      <td>8.210671</td>\n",
       "      <td>7.556953</td>\n",
       "      <td>7.485829</td>\n",
       "      <td>7.570994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZXDC</th>\n",
       "      <td>7.572224</td>\n",
       "      <td>7.815651</td>\n",
       "      <td>7.703752</td>\n",
       "      <td>7.813440</td>\n",
       "      <td>7.582254</td>\n",
       "      <td>7.848266</td>\n",
       "      <td>7.797961</td>\n",
       "      <td>7.747218</td>\n",
       "      <td>7.361969</td>\n",
       "      <td>7.762613</td>\n",
       "      <td>...</td>\n",
       "      <td>7.620372</td>\n",
       "      <td>7.481894</td>\n",
       "      <td>7.678229</td>\n",
       "      <td>7.501452</td>\n",
       "      <td>7.806311</td>\n",
       "      <td>7.756202</td>\n",
       "      <td>7.926744</td>\n",
       "      <td>7.596680</td>\n",
       "      <td>7.626171</td>\n",
       "      <td>7.542151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>8.298141</td>\n",
       "      <td>7.949561</td>\n",
       "      <td>8.062959</td>\n",
       "      <td>8.050516</td>\n",
       "      <td>7.910749</td>\n",
       "      <td>7.929676</td>\n",
       "      <td>7.618351</td>\n",
       "      <td>8.182171</td>\n",
       "      <td>7.580375</td>\n",
       "      <td>8.054994</td>\n",
       "      <td>...</td>\n",
       "      <td>7.658377</td>\n",
       "      <td>7.851989</td>\n",
       "      <td>7.911981</td>\n",
       "      <td>8.111525</td>\n",
       "      <td>7.727304</td>\n",
       "      <td>7.706939</td>\n",
       "      <td>7.725783</td>\n",
       "      <td>7.016313</td>\n",
       "      <td>7.102838</td>\n",
       "      <td>7.162837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>12.645891</td>\n",
       "      <td>12.748076</td>\n",
       "      <td>13.040601</td>\n",
       "      <td>12.692332</td>\n",
       "      <td>12.350203</td>\n",
       "      <td>12.544264</td>\n",
       "      <td>12.334911</td>\n",
       "      <td>12.968307</td>\n",
       "      <td>12.149615</td>\n",
       "      <td>13.021498</td>\n",
       "      <td>...</td>\n",
       "      <td>11.779425</td>\n",
       "      <td>12.431885</td>\n",
       "      <td>12.749677</td>\n",
       "      <td>12.808998</td>\n",
       "      <td>12.525400</td>\n",
       "      <td>12.953699</td>\n",
       "      <td>12.812301</td>\n",
       "      <td>12.754421</td>\n",
       "      <td>12.567707</td>\n",
       "      <td>12.586289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>8.953323</td>\n",
       "      <td>9.166833</td>\n",
       "      <td>9.169048</td>\n",
       "      <td>9.190854</td>\n",
       "      <td>9.024770</td>\n",
       "      <td>9.117730</td>\n",
       "      <td>9.261938</td>\n",
       "      <td>9.183532</td>\n",
       "      <td>9.194326</td>\n",
       "      <td>9.111614</td>\n",
       "      <td>...</td>\n",
       "      <td>9.304085</td>\n",
       "      <td>9.035303</td>\n",
       "      <td>8.888490</td>\n",
       "      <td>9.281351</td>\n",
       "      <td>9.262956</td>\n",
       "      <td>9.308908</td>\n",
       "      <td>9.143086</td>\n",
       "      <td>8.965400</td>\n",
       "      <td>9.133057</td>\n",
       "      <td>9.089495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>8.400813</td>\n",
       "      <td>8.230394</td>\n",
       "      <td>8.371081</td>\n",
       "      <td>8.241429</td>\n",
       "      <td>8.380828</td>\n",
       "      <td>8.169878</td>\n",
       "      <td>8.391880</td>\n",
       "      <td>8.324787</td>\n",
       "      <td>8.091579</td>\n",
       "      <td>8.420321</td>\n",
       "      <td>...</td>\n",
       "      <td>8.379961</td>\n",
       "      <td>8.507756</td>\n",
       "      <td>8.427223</td>\n",
       "      <td>8.381720</td>\n",
       "      <td>8.216812</td>\n",
       "      <td>8.320800</td>\n",
       "      <td>8.242512</td>\n",
       "      <td>7.787969</td>\n",
       "      <td>8.127396</td>\n",
       "      <td>8.073146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12268 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM6236086  GSM6236087  GSM6236088  GSM6236089  GSM6236090  \\\n",
       "A1BG        8.396817    8.540113    8.272503    8.261481    8.038747   \n",
       "A1BG-AS1    8.569792    8.566561    8.401626    8.390599    8.284091   \n",
       "A2M        15.252783   15.275463   15.530561   15.455341   15.120197   \n",
       "A4GALT     11.209895   11.128317   11.365849   10.987619   11.215685   \n",
       "AAAS        8.399242    8.117124    8.307808    8.230821    8.149916   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZXDC        7.572224    7.815651    7.703752    7.813440    7.582254   \n",
       "ZYG11B      8.298141    7.949561    8.062959    8.050516    7.910749   \n",
       "ZYX        12.645891   12.748076   13.040601   12.692332   12.350203   \n",
       "ZZEF1       8.953323    9.166833    9.169048    9.190854    9.024770   \n",
       "ZZZ3        8.400813    8.230394    8.371081    8.241429    8.380828   \n",
       "\n",
       "          GSM6236091  GSM6236092  GSM6236093  GSM6236094  GSM6236095  ...  \\\n",
       "A1BG        8.073204    8.005605    8.016628    7.865073    8.250083  ...   \n",
       "A1BG-AS1    8.243293    8.365375    8.403580    8.628567    8.536636  ...   \n",
       "A2M        15.357978   14.597179   15.539549   15.079838   15.541395  ...   \n",
       "A4GALT     10.982720   11.058364   11.125532   11.145937   10.993670  ...   \n",
       "AAAS        8.197217    8.146950    8.126790    7.851124    8.137845  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZXDC        7.848266    7.797961    7.747218    7.361969    7.762613  ...   \n",
       "ZYG11B      7.929676    7.618351    8.182171    7.580375    8.054994  ...   \n",
       "ZYX        12.544264   12.334911   12.968307   12.149615   13.021498  ...   \n",
       "ZZEF1       9.117730    9.261938    9.183532    9.194326    9.111614  ...   \n",
       "ZZZ3        8.169878    8.391880    8.324787    8.091579    8.420321  ...   \n",
       "\n",
       "          GSM6236106  GSM6236107  GSM6236108  GSM6236109  GSM6236110  \\\n",
       "A1BG        8.218620    8.115851    8.224858    8.079500    7.907925   \n",
       "A1BG-AS1    8.360774    8.144956    8.126360    8.305530    7.940901   \n",
       "A2M        14.793671   15.200568   15.416836   15.520590   14.606018   \n",
       "A4GALT     11.404432   11.664751   11.584156   11.537469   12.564070   \n",
       "AAAS        8.040290    8.132457    8.359567    7.994825    8.121188   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZXDC        7.620372    7.481894    7.678229    7.501452    7.806311   \n",
       "ZYG11B      7.658377    7.851989    7.911981    8.111525    7.727304   \n",
       "ZYX        11.779425   12.431885   12.749677   12.808998   12.525400   \n",
       "ZZEF1       9.304085    9.035303    8.888490    9.281351    9.262956   \n",
       "ZZZ3        8.379961    8.507756    8.427223    8.381720    8.216812   \n",
       "\n",
       "          GSM6236111  GSM6236112  GSM6236113  GSM6236114  GSM6236115  \n",
       "A1BG        8.315434    7.946120    7.287691    7.422345    7.434011  \n",
       "A1BG-AS1    8.089318    8.105210    8.017876    7.921990    7.906390  \n",
       "A2M        14.731785   14.648498   14.460178   14.779647   14.665974  \n",
       "A4GALT     12.512469   12.329296   12.028234   12.322689   12.130088  \n",
       "AAAS        8.262173    8.210671    7.556953    7.485829    7.570994  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "ZXDC        7.756202    7.926744    7.596680    7.626171    7.542151  \n",
       "ZYG11B      7.706939    7.725783    7.016313    7.102838    7.162837  \n",
       "ZYX        12.953699   12.812301   12.754421   12.567707   12.586289  \n",
       "ZZEF1       9.308908    9.143086    8.965400    9.133057    9.089495  \n",
       "ZZZ3        8.320800    8.242512    7.787969    8.127396    8.073146  \n",
       "\n",
       "[12268 rows x 30 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Poor Metabolizer of Drugs</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADACL3</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM6236086</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.396817</td>\n",
       "      <td>8.569792</td>\n",
       "      <td>15.252783</td>\n",
       "      <td>11.209895</td>\n",
       "      <td>8.399242</td>\n",
       "      <td>7.305539</td>\n",
       "      <td>6.753328</td>\n",
       "      <td>11.262105</td>\n",
       "      <td>7.958863</td>\n",
       "      <td>...</td>\n",
       "      <td>8.728034</td>\n",
       "      <td>8.763572</td>\n",
       "      <td>7.950313</td>\n",
       "      <td>10.027346</td>\n",
       "      <td>8.544385</td>\n",
       "      <td>7.572224</td>\n",
       "      <td>8.298141</td>\n",
       "      <td>12.645891</td>\n",
       "      <td>8.953323</td>\n",
       "      <td>8.400813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236087</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.540113</td>\n",
       "      <td>8.566561</td>\n",
       "      <td>15.275463</td>\n",
       "      <td>11.128317</td>\n",
       "      <td>8.117124</td>\n",
       "      <td>7.243744</td>\n",
       "      <td>7.291247</td>\n",
       "      <td>11.019643</td>\n",
       "      <td>8.362521</td>\n",
       "      <td>...</td>\n",
       "      <td>8.533005</td>\n",
       "      <td>8.472048</td>\n",
       "      <td>7.216584</td>\n",
       "      <td>8.747122</td>\n",
       "      <td>8.543741</td>\n",
       "      <td>7.815651</td>\n",
       "      <td>7.949561</td>\n",
       "      <td>12.748076</td>\n",
       "      <td>9.166833</td>\n",
       "      <td>8.230394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236088</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.272503</td>\n",
       "      <td>8.401626</td>\n",
       "      <td>15.530561</td>\n",
       "      <td>11.365849</td>\n",
       "      <td>8.307808</td>\n",
       "      <td>7.128770</td>\n",
       "      <td>6.706438</td>\n",
       "      <td>11.096122</td>\n",
       "      <td>8.167630</td>\n",
       "      <td>...</td>\n",
       "      <td>8.810304</td>\n",
       "      <td>8.438099</td>\n",
       "      <td>7.459568</td>\n",
       "      <td>9.173998</td>\n",
       "      <td>8.407976</td>\n",
       "      <td>7.703752</td>\n",
       "      <td>8.062959</td>\n",
       "      <td>13.040601</td>\n",
       "      <td>9.169048</td>\n",
       "      <td>8.371081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236089</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.261481</td>\n",
       "      <td>8.390599</td>\n",
       "      <td>15.455341</td>\n",
       "      <td>10.987619</td>\n",
       "      <td>8.230821</td>\n",
       "      <td>7.077062</td>\n",
       "      <td>7.117730</td>\n",
       "      <td>10.932092</td>\n",
       "      <td>8.395605</td>\n",
       "      <td>...</td>\n",
       "      <td>8.585088</td>\n",
       "      <td>8.535794</td>\n",
       "      <td>7.735612</td>\n",
       "      <td>9.793004</td>\n",
       "      <td>8.342001</td>\n",
       "      <td>7.813440</td>\n",
       "      <td>8.050516</td>\n",
       "      <td>12.692332</td>\n",
       "      <td>9.190854</td>\n",
       "      <td>8.241429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236090</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.038747</td>\n",
       "      <td>8.284091</td>\n",
       "      <td>15.120197</td>\n",
       "      <td>11.215685</td>\n",
       "      <td>8.149916</td>\n",
       "      <td>7.096870</td>\n",
       "      <td>6.975752</td>\n",
       "      <td>10.759791</td>\n",
       "      <td>7.933960</td>\n",
       "      <td>...</td>\n",
       "      <td>8.753272</td>\n",
       "      <td>8.381903</td>\n",
       "      <td>7.522647</td>\n",
       "      <td>8.954867</td>\n",
       "      <td>8.547473</td>\n",
       "      <td>7.582254</td>\n",
       "      <td>7.910749</td>\n",
       "      <td>12.350203</td>\n",
       "      <td>9.024770</td>\n",
       "      <td>8.380828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236091</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.073204</td>\n",
       "      <td>8.243293</td>\n",
       "      <td>15.357978</td>\n",
       "      <td>10.982720</td>\n",
       "      <td>8.197217</td>\n",
       "      <td>6.857150</td>\n",
       "      <td>7.243015</td>\n",
       "      <td>10.777899</td>\n",
       "      <td>8.396562</td>\n",
       "      <td>...</td>\n",
       "      <td>8.503097</td>\n",
       "      <td>8.297260</td>\n",
       "      <td>7.705863</td>\n",
       "      <td>9.819942</td>\n",
       "      <td>8.492921</td>\n",
       "      <td>7.848266</td>\n",
       "      <td>7.929676</td>\n",
       "      <td>12.544264</td>\n",
       "      <td>9.117730</td>\n",
       "      <td>8.169878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236092</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.005605</td>\n",
       "      <td>8.365375</td>\n",
       "      <td>14.597179</td>\n",
       "      <td>11.058364</td>\n",
       "      <td>8.146950</td>\n",
       "      <td>7.035064</td>\n",
       "      <td>7.393462</td>\n",
       "      <td>10.622494</td>\n",
       "      <td>8.300317</td>\n",
       "      <td>...</td>\n",
       "      <td>8.850109</td>\n",
       "      <td>8.705604</td>\n",
       "      <td>7.508851</td>\n",
       "      <td>9.008069</td>\n",
       "      <td>8.377500</td>\n",
       "      <td>7.797961</td>\n",
       "      <td>7.618351</td>\n",
       "      <td>12.334911</td>\n",
       "      <td>9.261938</td>\n",
       "      <td>8.391880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236093</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.016628</td>\n",
       "      <td>8.403580</td>\n",
       "      <td>15.539549</td>\n",
       "      <td>11.125532</td>\n",
       "      <td>8.126790</td>\n",
       "      <td>7.200875</td>\n",
       "      <td>7.205223</td>\n",
       "      <td>10.604708</td>\n",
       "      <td>8.426614</td>\n",
       "      <td>...</td>\n",
       "      <td>8.499779</td>\n",
       "      <td>8.197217</td>\n",
       "      <td>7.462025</td>\n",
       "      <td>9.061231</td>\n",
       "      <td>8.429686</td>\n",
       "      <td>7.747218</td>\n",
       "      <td>8.182171</td>\n",
       "      <td>12.968307</td>\n",
       "      <td>9.183532</td>\n",
       "      <td>8.324787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236094</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.865073</td>\n",
       "      <td>8.628567</td>\n",
       "      <td>15.079838</td>\n",
       "      <td>11.145937</td>\n",
       "      <td>7.851124</td>\n",
       "      <td>7.316431</td>\n",
       "      <td>7.308187</td>\n",
       "      <td>10.414395</td>\n",
       "      <td>8.062817</td>\n",
       "      <td>...</td>\n",
       "      <td>7.898853</td>\n",
       "      <td>8.017736</td>\n",
       "      <td>6.798094</td>\n",
       "      <td>7.711414</td>\n",
       "      <td>8.425216</td>\n",
       "      <td>7.361969</td>\n",
       "      <td>7.580375</td>\n",
       "      <td>12.149615</td>\n",
       "      <td>9.194326</td>\n",
       "      <td>8.091579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236095</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.250083</td>\n",
       "      <td>8.536636</td>\n",
       "      <td>15.541395</td>\n",
       "      <td>10.993670</td>\n",
       "      <td>8.137845</td>\n",
       "      <td>7.407584</td>\n",
       "      <td>6.985652</td>\n",
       "      <td>11.150826</td>\n",
       "      <td>8.242478</td>\n",
       "      <td>...</td>\n",
       "      <td>8.648477</td>\n",
       "      <td>8.587934</td>\n",
       "      <td>7.416023</td>\n",
       "      <td>8.732378</td>\n",
       "      <td>8.612684</td>\n",
       "      <td>7.762613</td>\n",
       "      <td>8.054994</td>\n",
       "      <td>13.021498</td>\n",
       "      <td>9.111614</td>\n",
       "      <td>8.420321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236096</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.306458</td>\n",
       "      <td>8.366031</td>\n",
       "      <td>15.674374</td>\n",
       "      <td>11.194367</td>\n",
       "      <td>8.280848</td>\n",
       "      <td>7.239594</td>\n",
       "      <td>6.757668</td>\n",
       "      <td>11.135581</td>\n",
       "      <td>8.097014</td>\n",
       "      <td>...</td>\n",
       "      <td>8.822279</td>\n",
       "      <td>8.576957</td>\n",
       "      <td>7.629721</td>\n",
       "      <td>9.269030</td>\n",
       "      <td>8.519964</td>\n",
       "      <td>7.583518</td>\n",
       "      <td>8.055749</td>\n",
       "      <td>13.028959</td>\n",
       "      <td>9.023154</td>\n",
       "      <td>8.450817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236097</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.328852</td>\n",
       "      <td>8.440835</td>\n",
       "      <td>15.360278</td>\n",
       "      <td>10.978067</td>\n",
       "      <td>8.307580</td>\n",
       "      <td>7.086270</td>\n",
       "      <td>7.109221</td>\n",
       "      <td>11.207111</td>\n",
       "      <td>8.438518</td>\n",
       "      <td>...</td>\n",
       "      <td>8.596128</td>\n",
       "      <td>8.258330</td>\n",
       "      <td>7.628263</td>\n",
       "      <td>9.646211</td>\n",
       "      <td>8.315263</td>\n",
       "      <td>7.747413</td>\n",
       "      <td>8.186613</td>\n",
       "      <td>12.443146</td>\n",
       "      <td>9.138186</td>\n",
       "      <td>8.004385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236098</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.384483</td>\n",
       "      <td>8.625313</td>\n",
       "      <td>14.439677</td>\n",
       "      <td>11.080718</td>\n",
       "      <td>7.989915</td>\n",
       "      <td>6.968702</td>\n",
       "      <td>7.359310</td>\n",
       "      <td>10.901986</td>\n",
       "      <td>8.006477</td>\n",
       "      <td>...</td>\n",
       "      <td>8.553501</td>\n",
       "      <td>8.096979</td>\n",
       "      <td>7.265849</td>\n",
       "      <td>8.390025</td>\n",
       "      <td>8.558293</td>\n",
       "      <td>7.709535</td>\n",
       "      <td>7.679642</td>\n",
       "      <td>11.448559</td>\n",
       "      <td>9.266435</td>\n",
       "      <td>8.087532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236099</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.144431</td>\n",
       "      <td>8.244602</td>\n",
       "      <td>15.406501</td>\n",
       "      <td>11.062254</td>\n",
       "      <td>8.367160</td>\n",
       "      <td>7.182036</td>\n",
       "      <td>7.026523</td>\n",
       "      <td>11.119325</td>\n",
       "      <td>8.294037</td>\n",
       "      <td>...</td>\n",
       "      <td>8.732676</td>\n",
       "      <td>8.467334</td>\n",
       "      <td>7.872829</td>\n",
       "      <td>9.845630</td>\n",
       "      <td>8.379884</td>\n",
       "      <td>7.838747</td>\n",
       "      <td>8.027988</td>\n",
       "      <td>12.761104</td>\n",
       "      <td>8.995885</td>\n",
       "      <td>8.243721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236100</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.380766</td>\n",
       "      <td>8.496055</td>\n",
       "      <td>14.386463</td>\n",
       "      <td>11.186981</td>\n",
       "      <td>8.507002</td>\n",
       "      <td>6.953085</td>\n",
       "      <td>7.103987</td>\n",
       "      <td>10.801284</td>\n",
       "      <td>8.158888</td>\n",
       "      <td>...</td>\n",
       "      <td>8.800522</td>\n",
       "      <td>8.679216</td>\n",
       "      <td>7.640847</td>\n",
       "      <td>9.194060</td>\n",
       "      <td>8.417466</td>\n",
       "      <td>7.370147</td>\n",
       "      <td>7.582117</td>\n",
       "      <td>12.039596</td>\n",
       "      <td>9.152581</td>\n",
       "      <td>8.269906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236101</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.904007</td>\n",
       "      <td>8.505878</td>\n",
       "      <td>15.250139</td>\n",
       "      <td>11.230371</td>\n",
       "      <td>7.874828</td>\n",
       "      <td>7.692065</td>\n",
       "      <td>7.207014</td>\n",
       "      <td>10.288847</td>\n",
       "      <td>7.960966</td>\n",
       "      <td>...</td>\n",
       "      <td>7.824641</td>\n",
       "      <td>8.283744</td>\n",
       "      <td>7.251167</td>\n",
       "      <td>7.794047</td>\n",
       "      <td>8.494922</td>\n",
       "      <td>7.415480</td>\n",
       "      <td>7.603170</td>\n",
       "      <td>12.610187</td>\n",
       "      <td>9.097944</td>\n",
       "      <td>8.382230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236102</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.220335</td>\n",
       "      <td>8.376415</td>\n",
       "      <td>15.301475</td>\n",
       "      <td>11.657746</td>\n",
       "      <td>8.350055</td>\n",
       "      <td>7.172497</td>\n",
       "      <td>7.313941</td>\n",
       "      <td>10.931409</td>\n",
       "      <td>8.326932</td>\n",
       "      <td>...</td>\n",
       "      <td>8.607392</td>\n",
       "      <td>8.492921</td>\n",
       "      <td>7.923327</td>\n",
       "      <td>9.852039</td>\n",
       "      <td>8.603348</td>\n",
       "      <td>7.927407</td>\n",
       "      <td>8.201115</td>\n",
       "      <td>12.599339</td>\n",
       "      <td>9.268639</td>\n",
       "      <td>8.322886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236103</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.256255</td>\n",
       "      <td>8.488978</td>\n",
       "      <td>15.495640</td>\n",
       "      <td>11.448559</td>\n",
       "      <td>8.278140</td>\n",
       "      <td>7.213345</td>\n",
       "      <td>7.033974</td>\n",
       "      <td>11.103364</td>\n",
       "      <td>8.262250</td>\n",
       "      <td>...</td>\n",
       "      <td>8.603874</td>\n",
       "      <td>8.469337</td>\n",
       "      <td>7.354617</td>\n",
       "      <td>8.827939</td>\n",
       "      <td>8.717848</td>\n",
       "      <td>7.759194</td>\n",
       "      <td>8.342477</td>\n",
       "      <td>12.849698</td>\n",
       "      <td>9.209108</td>\n",
       "      <td>8.394934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236104</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.656103</td>\n",
       "      <td>8.287751</td>\n",
       "      <td>15.210812</td>\n",
       "      <td>11.452696</td>\n",
       "      <td>7.901671</td>\n",
       "      <td>7.052445</td>\n",
       "      <td>7.203511</td>\n",
       "      <td>11.194665</td>\n",
       "      <td>8.479464</td>\n",
       "      <td>...</td>\n",
       "      <td>8.663647</td>\n",
       "      <td>8.196930</td>\n",
       "      <td>7.098559</td>\n",
       "      <td>8.967662</td>\n",
       "      <td>8.287751</td>\n",
       "      <td>7.595811</td>\n",
       "      <td>8.054921</td>\n",
       "      <td>12.474496</td>\n",
       "      <td>9.214360</td>\n",
       "      <td>8.041451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236105</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.450492</td>\n",
       "      <td>8.198445</td>\n",
       "      <td>15.569610</td>\n",
       "      <td>11.506812</td>\n",
       "      <td>8.344444</td>\n",
       "      <td>7.149039</td>\n",
       "      <td>6.931919</td>\n",
       "      <td>10.405921</td>\n",
       "      <td>8.130371</td>\n",
       "      <td>...</td>\n",
       "      <td>8.910893</td>\n",
       "      <td>8.537510</td>\n",
       "      <td>7.956134</td>\n",
       "      <td>9.944092</td>\n",
       "      <td>8.532810</td>\n",
       "      <td>7.630311</td>\n",
       "      <td>8.138143</td>\n",
       "      <td>12.670173</td>\n",
       "      <td>8.823288</td>\n",
       "      <td>8.476353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236106</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.218620</td>\n",
       "      <td>8.360774</td>\n",
       "      <td>14.793671</td>\n",
       "      <td>11.404432</td>\n",
       "      <td>8.040290</td>\n",
       "      <td>7.142201</td>\n",
       "      <td>7.063934</td>\n",
       "      <td>10.248046</td>\n",
       "      <td>7.986808</td>\n",
       "      <td>...</td>\n",
       "      <td>8.764927</td>\n",
       "      <td>8.199918</td>\n",
       "      <td>7.454642</td>\n",
       "      <td>8.552930</td>\n",
       "      <td>8.751321</td>\n",
       "      <td>7.620372</td>\n",
       "      <td>7.658377</td>\n",
       "      <td>11.779425</td>\n",
       "      <td>9.304085</td>\n",
       "      <td>8.379961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236107</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.115851</td>\n",
       "      <td>8.144956</td>\n",
       "      <td>15.200568</td>\n",
       "      <td>11.664751</td>\n",
       "      <td>8.132457</td>\n",
       "      <td>7.179357</td>\n",
       "      <td>6.861294</td>\n",
       "      <td>9.882121</td>\n",
       "      <td>7.909237</td>\n",
       "      <td>...</td>\n",
       "      <td>8.650394</td>\n",
       "      <td>8.207624</td>\n",
       "      <td>7.678013</td>\n",
       "      <td>8.920658</td>\n",
       "      <td>8.778734</td>\n",
       "      <td>7.481894</td>\n",
       "      <td>7.851989</td>\n",
       "      <td>12.431885</td>\n",
       "      <td>9.035303</td>\n",
       "      <td>8.507756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236108</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.224858</td>\n",
       "      <td>8.126360</td>\n",
       "      <td>15.416836</td>\n",
       "      <td>11.584156</td>\n",
       "      <td>8.359567</td>\n",
       "      <td>7.132847</td>\n",
       "      <td>6.956521</td>\n",
       "      <td>10.526002</td>\n",
       "      <td>8.022059</td>\n",
       "      <td>...</td>\n",
       "      <td>8.739190</td>\n",
       "      <td>8.496321</td>\n",
       "      <td>7.885646</td>\n",
       "      <td>10.119259</td>\n",
       "      <td>8.445360</td>\n",
       "      <td>7.678229</td>\n",
       "      <td>7.911981</td>\n",
       "      <td>12.749677</td>\n",
       "      <td>8.888490</td>\n",
       "      <td>8.427223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236109</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.079500</td>\n",
       "      <td>8.305530</td>\n",
       "      <td>15.520590</td>\n",
       "      <td>11.537469</td>\n",
       "      <td>7.994825</td>\n",
       "      <td>7.287341</td>\n",
       "      <td>6.947393</td>\n",
       "      <td>10.179069</td>\n",
       "      <td>8.247282</td>\n",
       "      <td>...</td>\n",
       "      <td>8.365010</td>\n",
       "      <td>8.120540</td>\n",
       "      <td>7.484756</td>\n",
       "      <td>9.155444</td>\n",
       "      <td>8.344333</td>\n",
       "      <td>7.501452</td>\n",
       "      <td>8.111525</td>\n",
       "      <td>12.808998</td>\n",
       "      <td>9.281351</td>\n",
       "      <td>8.381720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236110</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.907925</td>\n",
       "      <td>7.940901</td>\n",
       "      <td>14.606018</td>\n",
       "      <td>12.564070</td>\n",
       "      <td>8.121188</td>\n",
       "      <td>6.764579</td>\n",
       "      <td>7.831413</td>\n",
       "      <td>10.465804</td>\n",
       "      <td>8.458384</td>\n",
       "      <td>...</td>\n",
       "      <td>8.457620</td>\n",
       "      <td>8.147544</td>\n",
       "      <td>7.095924</td>\n",
       "      <td>8.422383</td>\n",
       "      <td>8.602637</td>\n",
       "      <td>7.806311</td>\n",
       "      <td>7.727304</td>\n",
       "      <td>12.525400</td>\n",
       "      <td>9.262956</td>\n",
       "      <td>8.216812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236111</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.315434</td>\n",
       "      <td>8.089318</td>\n",
       "      <td>14.731785</td>\n",
       "      <td>12.512469</td>\n",
       "      <td>8.262173</td>\n",
       "      <td>7.026054</td>\n",
       "      <td>7.845229</td>\n",
       "      <td>10.414544</td>\n",
       "      <td>8.408393</td>\n",
       "      <td>...</td>\n",
       "      <td>8.312845</td>\n",
       "      <td>8.406205</td>\n",
       "      <td>7.253296</td>\n",
       "      <td>8.422981</td>\n",
       "      <td>8.683141</td>\n",
       "      <td>7.756202</td>\n",
       "      <td>7.706939</td>\n",
       "      <td>12.953699</td>\n",
       "      <td>9.308908</td>\n",
       "      <td>8.320800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236112</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.946120</td>\n",
       "      <td>8.105210</td>\n",
       "      <td>14.648498</td>\n",
       "      <td>12.329296</td>\n",
       "      <td>8.210671</td>\n",
       "      <td>6.937300</td>\n",
       "      <td>7.932510</td>\n",
       "      <td>10.574956</td>\n",
       "      <td>8.448592</td>\n",
       "      <td>...</td>\n",
       "      <td>8.519964</td>\n",
       "      <td>8.253926</td>\n",
       "      <td>7.263347</td>\n",
       "      <td>8.411563</td>\n",
       "      <td>8.441976</td>\n",
       "      <td>7.926744</td>\n",
       "      <td>7.725783</td>\n",
       "      <td>12.812301</td>\n",
       "      <td>9.143086</td>\n",
       "      <td>8.242512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236113</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.287691</td>\n",
       "      <td>8.017876</td>\n",
       "      <td>14.460178</td>\n",
       "      <td>12.028234</td>\n",
       "      <td>7.556953</td>\n",
       "      <td>7.287091</td>\n",
       "      <td>8.114003</td>\n",
       "      <td>10.173656</td>\n",
       "      <td>8.686097</td>\n",
       "      <td>...</td>\n",
       "      <td>7.415671</td>\n",
       "      <td>7.601028</td>\n",
       "      <td>6.863980</td>\n",
       "      <td>7.103115</td>\n",
       "      <td>8.302601</td>\n",
       "      <td>7.596680</td>\n",
       "      <td>7.016313</td>\n",
       "      <td>12.754421</td>\n",
       "      <td>8.965400</td>\n",
       "      <td>7.787969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236114</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.422345</td>\n",
       "      <td>7.921990</td>\n",
       "      <td>14.779647</td>\n",
       "      <td>12.322689</td>\n",
       "      <td>7.485829</td>\n",
       "      <td>7.430965</td>\n",
       "      <td>7.659996</td>\n",
       "      <td>9.926296</td>\n",
       "      <td>8.356510</td>\n",
       "      <td>...</td>\n",
       "      <td>7.570109</td>\n",
       "      <td>7.727183</td>\n",
       "      <td>6.929653</td>\n",
       "      <td>7.415918</td>\n",
       "      <td>8.518358</td>\n",
       "      <td>7.626171</td>\n",
       "      <td>7.102838</td>\n",
       "      <td>12.567707</td>\n",
       "      <td>9.133057</td>\n",
       "      <td>8.127396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6236115</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.434011</td>\n",
       "      <td>7.906390</td>\n",
       "      <td>14.665974</td>\n",
       "      <td>12.130088</td>\n",
       "      <td>7.570994</td>\n",
       "      <td>7.318596</td>\n",
       "      <td>8.000939</td>\n",
       "      <td>10.483749</td>\n",
       "      <td>8.388690</td>\n",
       "      <td>...</td>\n",
       "      <td>7.492120</td>\n",
       "      <td>7.991333</td>\n",
       "      <td>7.002628</td>\n",
       "      <td>7.417467</td>\n",
       "      <td>8.442252</td>\n",
       "      <td>7.542151</td>\n",
       "      <td>7.162837</td>\n",
       "      <td>12.586289</td>\n",
       "      <td>9.089495</td>\n",
       "      <td>8.073146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 12269 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Poor Metabolizer of Drugs      A1BG  A1BG-AS1        A2M  \\\n",
       "GSM6236086                        0.0  8.396817  8.569792  15.252783   \n",
       "GSM6236087                        0.0  8.540113  8.566561  15.275463   \n",
       "GSM6236088                        0.0  8.272503  8.401626  15.530561   \n",
       "GSM6236089                        0.0  8.261481  8.390599  15.455341   \n",
       "GSM6236090                        1.0  8.038747  8.284091  15.120197   \n",
       "GSM6236091                        1.0  8.073204  8.243293  15.357978   \n",
       "GSM6236092                        0.0  8.005605  8.365375  14.597179   \n",
       "GSM6236093                        0.0  8.016628  8.403580  15.539549   \n",
       "GSM6236094                        0.0  7.865073  8.628567  15.079838   \n",
       "GSM6236095                        0.0  8.250083  8.536636  15.541395   \n",
       "GSM6236096                        0.0  8.306458  8.366031  15.674374   \n",
       "GSM6236097                        0.0  8.328852  8.440835  15.360278   \n",
       "GSM6236098                        1.0  8.384483  8.625313  14.439677   \n",
       "GSM6236099                        1.0  8.144431  8.244602  15.406501   \n",
       "GSM6236100                        0.0  8.380766  8.496055  14.386463   \n",
       "GSM6236101                        0.0  7.904007  8.505878  15.250139   \n",
       "GSM6236102                        0.0  8.220335  8.376415  15.301475   \n",
       "GSM6236103                        0.0  8.256255  8.488978  15.495640   \n",
       "GSM6236104                        0.0  8.656103  8.287751  15.210812   \n",
       "GSM6236105                        0.0  8.450492  8.198445  15.569610   \n",
       "GSM6236106                        1.0  8.218620  8.360774  14.793671   \n",
       "GSM6236107                        1.0  8.115851  8.144956  15.200568   \n",
       "GSM6236108                        1.0  8.224858  8.126360  15.416836   \n",
       "GSM6236109                        0.0  8.079500  8.305530  15.520590   \n",
       "GSM6236110                        0.0  7.907925  7.940901  14.606018   \n",
       "GSM6236111                        0.0  8.315434  8.089318  14.731785   \n",
       "GSM6236112                        0.0  7.946120  8.105210  14.648498   \n",
       "GSM6236113                        0.0  7.287691  8.017876  14.460178   \n",
       "GSM6236114                        0.0  7.422345  7.921990  14.779647   \n",
       "GSM6236115                        0.0  7.434011  7.906390  14.665974   \n",
       "\n",
       "               A4GALT      AAAS      AACS   AADACL3      AAGAB      AAK1  ...  \\\n",
       "GSM6236086  11.209895  8.399242  7.305539  6.753328  11.262105  7.958863  ...   \n",
       "GSM6236087  11.128317  8.117124  7.243744  7.291247  11.019643  8.362521  ...   \n",
       "GSM6236088  11.365849  8.307808  7.128770  6.706438  11.096122  8.167630  ...   \n",
       "GSM6236089  10.987619  8.230821  7.077062  7.117730  10.932092  8.395605  ...   \n",
       "GSM6236090  11.215685  8.149916  7.096870  6.975752  10.759791  7.933960  ...   \n",
       "GSM6236091  10.982720  8.197217  6.857150  7.243015  10.777899  8.396562  ...   \n",
       "GSM6236092  11.058364  8.146950  7.035064  7.393462  10.622494  8.300317  ...   \n",
       "GSM6236093  11.125532  8.126790  7.200875  7.205223  10.604708  8.426614  ...   \n",
       "GSM6236094  11.145937  7.851124  7.316431  7.308187  10.414395  8.062817  ...   \n",
       "GSM6236095  10.993670  8.137845  7.407584  6.985652  11.150826  8.242478  ...   \n",
       "GSM6236096  11.194367  8.280848  7.239594  6.757668  11.135581  8.097014  ...   \n",
       "GSM6236097  10.978067  8.307580  7.086270  7.109221  11.207111  8.438518  ...   \n",
       "GSM6236098  11.080718  7.989915  6.968702  7.359310  10.901986  8.006477  ...   \n",
       "GSM6236099  11.062254  8.367160  7.182036  7.026523  11.119325  8.294037  ...   \n",
       "GSM6236100  11.186981  8.507002  6.953085  7.103987  10.801284  8.158888  ...   \n",
       "GSM6236101  11.230371  7.874828  7.692065  7.207014  10.288847  7.960966  ...   \n",
       "GSM6236102  11.657746  8.350055  7.172497  7.313941  10.931409  8.326932  ...   \n",
       "GSM6236103  11.448559  8.278140  7.213345  7.033974  11.103364  8.262250  ...   \n",
       "GSM6236104  11.452696  7.901671  7.052445  7.203511  11.194665  8.479464  ...   \n",
       "GSM6236105  11.506812  8.344444  7.149039  6.931919  10.405921  8.130371  ...   \n",
       "GSM6236106  11.404432  8.040290  7.142201  7.063934  10.248046  7.986808  ...   \n",
       "GSM6236107  11.664751  8.132457  7.179357  6.861294   9.882121  7.909237  ...   \n",
       "GSM6236108  11.584156  8.359567  7.132847  6.956521  10.526002  8.022059  ...   \n",
       "GSM6236109  11.537469  7.994825  7.287341  6.947393  10.179069  8.247282  ...   \n",
       "GSM6236110  12.564070  8.121188  6.764579  7.831413  10.465804  8.458384  ...   \n",
       "GSM6236111  12.512469  8.262173  7.026054  7.845229  10.414544  8.408393  ...   \n",
       "GSM6236112  12.329296  8.210671  6.937300  7.932510  10.574956  8.448592  ...   \n",
       "GSM6236113  12.028234  7.556953  7.287091  8.114003  10.173656  8.686097  ...   \n",
       "GSM6236114  12.322689  7.485829  7.430965  7.659996   9.926296  8.356510  ...   \n",
       "GSM6236115  12.130088  7.570994  7.318596  8.000939  10.483749  8.388690  ...   \n",
       "\n",
       "              ZSWIM7      ZW10    ZWILCH      ZWINT      ZXDB      ZXDC  \\\n",
       "GSM6236086  8.728034  8.763572  7.950313  10.027346  8.544385  7.572224   \n",
       "GSM6236087  8.533005  8.472048  7.216584   8.747122  8.543741  7.815651   \n",
       "GSM6236088  8.810304  8.438099  7.459568   9.173998  8.407976  7.703752   \n",
       "GSM6236089  8.585088  8.535794  7.735612   9.793004  8.342001  7.813440   \n",
       "GSM6236090  8.753272  8.381903  7.522647   8.954867  8.547473  7.582254   \n",
       "GSM6236091  8.503097  8.297260  7.705863   9.819942  8.492921  7.848266   \n",
       "GSM6236092  8.850109  8.705604  7.508851   9.008069  8.377500  7.797961   \n",
       "GSM6236093  8.499779  8.197217  7.462025   9.061231  8.429686  7.747218   \n",
       "GSM6236094  7.898853  8.017736  6.798094   7.711414  8.425216  7.361969   \n",
       "GSM6236095  8.648477  8.587934  7.416023   8.732378  8.612684  7.762613   \n",
       "GSM6236096  8.822279  8.576957  7.629721   9.269030  8.519964  7.583518   \n",
       "GSM6236097  8.596128  8.258330  7.628263   9.646211  8.315263  7.747413   \n",
       "GSM6236098  8.553501  8.096979  7.265849   8.390025  8.558293  7.709535   \n",
       "GSM6236099  8.732676  8.467334  7.872829   9.845630  8.379884  7.838747   \n",
       "GSM6236100  8.800522  8.679216  7.640847   9.194060  8.417466  7.370147   \n",
       "GSM6236101  7.824641  8.283744  7.251167   7.794047  8.494922  7.415480   \n",
       "GSM6236102  8.607392  8.492921  7.923327   9.852039  8.603348  7.927407   \n",
       "GSM6236103  8.603874  8.469337  7.354617   8.827939  8.717848  7.759194   \n",
       "GSM6236104  8.663647  8.196930  7.098559   8.967662  8.287751  7.595811   \n",
       "GSM6236105  8.910893  8.537510  7.956134   9.944092  8.532810  7.630311   \n",
       "GSM6236106  8.764927  8.199918  7.454642   8.552930  8.751321  7.620372   \n",
       "GSM6236107  8.650394  8.207624  7.678013   8.920658  8.778734  7.481894   \n",
       "GSM6236108  8.739190  8.496321  7.885646  10.119259  8.445360  7.678229   \n",
       "GSM6236109  8.365010  8.120540  7.484756   9.155444  8.344333  7.501452   \n",
       "GSM6236110  8.457620  8.147544  7.095924   8.422383  8.602637  7.806311   \n",
       "GSM6236111  8.312845  8.406205  7.253296   8.422981  8.683141  7.756202   \n",
       "GSM6236112  8.519964  8.253926  7.263347   8.411563  8.441976  7.926744   \n",
       "GSM6236113  7.415671  7.601028  6.863980   7.103115  8.302601  7.596680   \n",
       "GSM6236114  7.570109  7.727183  6.929653   7.415918  8.518358  7.626171   \n",
       "GSM6236115  7.492120  7.991333  7.002628   7.417467  8.442252  7.542151   \n",
       "\n",
       "              ZYG11B        ZYX     ZZEF1      ZZZ3  \n",
       "GSM6236086  8.298141  12.645891  8.953323  8.400813  \n",
       "GSM6236087  7.949561  12.748076  9.166833  8.230394  \n",
       "GSM6236088  8.062959  13.040601  9.169048  8.371081  \n",
       "GSM6236089  8.050516  12.692332  9.190854  8.241429  \n",
       "GSM6236090  7.910749  12.350203  9.024770  8.380828  \n",
       "GSM6236091  7.929676  12.544264  9.117730  8.169878  \n",
       "GSM6236092  7.618351  12.334911  9.261938  8.391880  \n",
       "GSM6236093  8.182171  12.968307  9.183532  8.324787  \n",
       "GSM6236094  7.580375  12.149615  9.194326  8.091579  \n",
       "GSM6236095  8.054994  13.021498  9.111614  8.420321  \n",
       "GSM6236096  8.055749  13.028959  9.023154  8.450817  \n",
       "GSM6236097  8.186613  12.443146  9.138186  8.004385  \n",
       "GSM6236098  7.679642  11.448559  9.266435  8.087532  \n",
       "GSM6236099  8.027988  12.761104  8.995885  8.243721  \n",
       "GSM6236100  7.582117  12.039596  9.152581  8.269906  \n",
       "GSM6236101  7.603170  12.610187  9.097944  8.382230  \n",
       "GSM6236102  8.201115  12.599339  9.268639  8.322886  \n",
       "GSM6236103  8.342477  12.849698  9.209108  8.394934  \n",
       "GSM6236104  8.054921  12.474496  9.214360  8.041451  \n",
       "GSM6236105  8.138143  12.670173  8.823288  8.476353  \n",
       "GSM6236106  7.658377  11.779425  9.304085  8.379961  \n",
       "GSM6236107  7.851989  12.431885  9.035303  8.507756  \n",
       "GSM6236108  7.911981  12.749677  8.888490  8.427223  \n",
       "GSM6236109  8.111525  12.808998  9.281351  8.381720  \n",
       "GSM6236110  7.727304  12.525400  9.262956  8.216812  \n",
       "GSM6236111  7.706939  12.953699  9.308908  8.320800  \n",
       "GSM6236112  7.725783  12.812301  9.143086  8.242512  \n",
       "GSM6236113  7.016313  12.754421  8.965400  7.787969  \n",
       "GSM6236114  7.102838  12.567707  9.133057  8.127396  \n",
       "GSM6236115  7.162837  12.586289  9.089495  8.073146  \n",
       "\n",
       "[30 rows x 12269 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 30 samples.\n",
      "For the feature 'Poor Metabolizer of Drugs', the least common label is '1.0' with 7 occurrences. This represents 23.33% of the dataset.\n",
      "The distribution of the feature 'Poor Metabolizer of Drugs' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output\\Jiayi\\Poor-Metabolizer-of-Drugs\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Comparative analysis of glioblastoma initiating cells and patient-matched EPSC-derived neural stem cells as a discovery tool and drug matching strategy [array]\"\n",
      "!Series_summary\t\"Malignant gliomas are the most common intrinsic brain tumors in adults and are associated with a very poor prognosis, which has not improved in the last 20 years. We show here that glioblastoma initiating cells (GIC) and syngeneic neural stem cells derived from enhanced pluripotent stem cells (iNSC) are a suitable novel pre-clinical pipeline to discover new patient-specific disease mechanisms and to identify druggable targets for personalized therapy.\"\n",
      "!Series_overall_design\t\"Methylation analysis of human dura mater fibroblasts, iPSC, iNSC, GIC and FFPE bulk tumor\"\n"
     ]
    }
   ],
   "source": [
    "import io\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM4718577</th>\n",
       "      <th>GSM4718578</th>\n",
       "      <th>GSM4718579</th>\n",
       "      <th>GSM4718580</th>\n",
       "      <th>GSM4718581</th>\n",
       "      <th>GSM4718582</th>\n",
       "      <th>GSM4718583</th>\n",
       "      <th>GSM4718584</th>\n",
       "      <th>GSM4718585</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4718659</th>\n",
       "      <th>GSM4718660</th>\n",
       "      <th>GSM4718661</th>\n",
       "      <th>GSM4718662</th>\n",
       "      <th>GSM4718663</th>\n",
       "      <th>GSM4718664</th>\n",
       "      <th>GSM4718665</th>\n",
       "      <th>GSM4718666</th>\n",
       "      <th>GSM4718667</th>\n",
       "      <th>GSM4718668</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>patient: 17</td>\n",
       "      <td>patient: 17</td>\n",
       "      <td>patient: 18</td>\n",
       "      <td>patient: 18</td>\n",
       "      <td>patient: 18</td>\n",
       "      <td>patient: 19</td>\n",
       "      <td>patient: 19</td>\n",
       "      <td>patient: 19</td>\n",
       "      <td>patient: 19</td>\n",
       "      <td>...</td>\n",
       "      <td>patient: 18</td>\n",
       "      <td>patient: 19</td>\n",
       "      <td>patient: 30</td>\n",
       "      <td>patient: 31</td>\n",
       "      <td>patient: 17</td>\n",
       "      <td>patient: 50</td>\n",
       "      <td>patient: 54</td>\n",
       "      <td>patient: 61</td>\n",
       "      <td>patient: 26</td>\n",
       "      <td>patient: 52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>passage: 7</td>\n",
       "      <td>passage: 4</td>\n",
       "      <td>passage: 6</td>\n",
       "      <td>passage: 6</td>\n",
       "      <td>passage: 4</td>\n",
       "      <td>passage: 7</td>\n",
       "      <td>passage: 13</td>\n",
       "      <td>passage: 2</td>\n",
       "      <td>passage: 2</td>\n",
       "      <td>...</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "      <td>passage: NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>sample type: cell_culture</td>\n",
       "      <td>...</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "      <td>sample type: FFPE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: FB</td>\n",
       "      <td>cell type: iNSC</td>\n",
       "      <td>cell type: FB</td>\n",
       "      <td>cell type: iNSC</td>\n",
       "      <td>cell type: iNSC</td>\n",
       "      <td>cell type: FB</td>\n",
       "      <td>cell type: iPSC</td>\n",
       "      <td>cell type: iNSC</td>\n",
       "      <td>cell type: iNSC</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "      <td>cell type: GBM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                 GSM4718577  \\\n",
       "0  !Sample_characteristics_ch1                patient: 17   \n",
       "1  !Sample_characteristics_ch1                 passage: 7   \n",
       "2  !Sample_characteristics_ch1  sample type: cell_culture   \n",
       "3  !Sample_characteristics_ch1              cell type: FB   \n",
       "\n",
       "                  GSM4718578                 GSM4718579  \\\n",
       "0                patient: 17                patient: 18   \n",
       "1                 passage: 4                 passage: 6   \n",
       "2  sample type: cell_culture  sample type: cell_culture   \n",
       "3            cell type: iNSC              cell type: FB   \n",
       "\n",
       "                  GSM4718580                 GSM4718581  \\\n",
       "0                patient: 18                patient: 18   \n",
       "1                 passage: 6                 passage: 4   \n",
       "2  sample type: cell_culture  sample type: cell_culture   \n",
       "3            cell type: iNSC            cell type: iNSC   \n",
       "\n",
       "                  GSM4718582                 GSM4718583  \\\n",
       "0                patient: 19                patient: 19   \n",
       "1                 passage: 7                passage: 13   \n",
       "2  sample type: cell_culture  sample type: cell_culture   \n",
       "3              cell type: FB            cell type: iPSC   \n",
       "\n",
       "                  GSM4718584                 GSM4718585  ...  \\\n",
       "0                patient: 19                patient: 19  ...   \n",
       "1                 passage: 2                 passage: 2  ...   \n",
       "2  sample type: cell_culture  sample type: cell_culture  ...   \n",
       "3            cell type: iNSC            cell type: iNSC  ...   \n",
       "\n",
       "          GSM4718659         GSM4718660         GSM4718661         GSM4718662  \\\n",
       "0        patient: 18        patient: 19        patient: 30        patient: 31   \n",
       "1        passage: NA        passage: NA        passage: NA        passage: NA   \n",
       "2  sample type: FFPE  sample type: FFPE  sample type: FFPE  sample type: FFPE   \n",
       "3     cell type: GBM     cell type: GBM     cell type: GBM     cell type: GBM   \n",
       "\n",
       "          GSM4718663         GSM4718664         GSM4718665         GSM4718666  \\\n",
       "0        patient: 17        patient: 50        patient: 54        patient: 61   \n",
       "1        passage: NA        passage: NA        passage: NA        passage: NA   \n",
       "2  sample type: FFPE  sample type: FFPE  sample type: FFPE  sample type: FFPE   \n",
       "3     cell type: GBM     cell type: GBM     cell type: GBM     cell type: GBM   \n",
       "\n",
       "          GSM4718667         GSM4718668  \n",
       "0        patient: 26        patient: 52  \n",
       "1        passage: NA        passage: NA  \n",
       "2  sample type: FFPE  sample type: FFPE  \n",
       "3     cell type: GBM     cell type: GBM  \n",
       "\n",
       "[4 rows x 63 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['patient: 17',\n",
       "  'patient: 18',\n",
       "  'patient: 19',\n",
       "  'patient: 26',\n",
       "  'patient: 30',\n",
       "  'patient: 31',\n",
       "  'patient: 50',\n",
       "  'patient: 52',\n",
       "  'patient: 54',\n",
       "  'patient: 61'],\n",
       " 1: ['passage: 7',\n",
       "  'passage: 4',\n",
       "  'passage: 6',\n",
       "  'passage: 13',\n",
       "  'passage: 2',\n",
       "  'passage: 8',\n",
       "  'passage: 5',\n",
       "  'passage: 1',\n",
       "  'passage: 10',\n",
       "  'passage: 3',\n",
       "  'passage: 11',\n",
       "  'passage: 3;4',\n",
       "  'passage: 12',\n",
       "  'passage: 3;6',\n",
       "  'passage: 9',\n",
       "  'passage: 7;8',\n",
       "  'passage: 4;5',\n",
       "  'passage: 6;7',\n",
       "  'passage: NA'],\n",
       " 2: ['sample type: cell_culture', 'sample type: FFPE'],\n",
       " 3: ['cell type: FB',\n",
       "  'cell type: iNSC',\n",
       "  'cell type: iPSC',\n",
       "  'cell type: GIC',\n",
       "  'cell type: GBM']}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Poor Metabolizer of Drugs\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Poor Metabolizer of Drugs\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Comparative analysis of glioblastoma initiating cells and patient-matched EPSC-derived neural stem cells as a discovery tool and drug matching strategy [array]\"\\n!Series_summary\\t\"Malignant gliomas are the most common intrinsic brain tumors in adults and are associated with a very poor prognosis, which has not improved in the last 20 years. We show here that glioblastoma initiating cells (GIC) and syngeneic neural stem cells derived from enhanced pluripotent stem cells (iNSC) are a suitable novel pre-clinical pipeline to discover new patient-specific disease mechanisms and to identify druggable targets for personalized therapy.\"\\n!Series_overall_design\\t\"Methylation analysis of human dura mater fibroblasts, iPSC, iNSC, GIC and FFPE bulk tumor\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'patient: 17\\', \\'patient: 18\\', \\'patient: 19\\', \\'patient: 26\\', \\'patient: 30\\', \\'patient: 31\\', \\'patient: 50\\', \\'patient: 52\\', \\'patient: 54\\', \\'patient: 61\\'], 1: [\\'passage: 7\\', \\'passage: 4\\', \\'passage: 6\\', \\'passage: 13\\', \\'passage: 2\\', \\'passage: 8\\', \\'passage: 5\\', \\'passage: 1\\', \\'passage: 10\\', \\'passage: 3\\', \\'passage: 11\\', \\'passage: 3;4\\', \\'passage: 12\\', \\'passage: 3;6\\', \\'passage: 9\\', \\'passage: 7;8\\', \\'passage: 4;5\\', \\'passage: 6;7\\', \\'passage: NA\\'], 2: [\\'sample type: cell_culture\\', \\'sample type: FFPE\\'], 3: [\\'cell type: FB\\', \\'cell type: iNSC\\', \\'cell type: iPSC\\', \\'cell type: GIC\\', \\'cell type: GBM\\']}\\n'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 2\n",
    "age_row = None\n",
    "gender_row = None\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Verify and use the functions generated by GPT\n",
    "\n",
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'sample type: cell_culture':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    Convert age string to a continuous numerical value.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        # Extract age as an integer from the string\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        # In case of any format error or unexpected string structure\n",
    "        return None\n",
    "\n",
    "\n",
    "# It sometimes maps 'female' to 0, and sometimes 1. Does it matter?\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary value.\n",
    "    'female' is represented as 1, 'male' as 0.\n",
    "    Unknown values are converted to None.\n",
    "    \"\"\"\n",
    "    if gender_string.lower() == 'male: 0':\n",
    "        return 0\n",
    "    elif gender_string.lower() == 'male: 1':\n",
    "        return 1\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM4718577</th>\n",
       "      <th>GSM4718578</th>\n",
       "      <th>GSM4718579</th>\n",
       "      <th>GSM4718580</th>\n",
       "      <th>GSM4718581</th>\n",
       "      <th>GSM4718582</th>\n",
       "      <th>GSM4718583</th>\n",
       "      <th>GSM4718584</th>\n",
       "      <th>GSM4718585</th>\n",
       "      <th>GSM4718586</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4718659</th>\n",
       "      <th>GSM4718660</th>\n",
       "      <th>GSM4718661</th>\n",
       "      <th>GSM4718662</th>\n",
       "      <th>GSM4718663</th>\n",
       "      <th>GSM4718664</th>\n",
       "      <th>GSM4718665</th>\n",
       "      <th>GSM4718666</th>\n",
       "      <th>GSM4718667</th>\n",
       "      <th>GSM4718668</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Poor Metabolizer of Drugs</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           GSM4718577  GSM4718578  GSM4718579  GSM4718580  \\\n",
       "Poor Metabolizer of Drugs           1           1           1           1   \n",
       "\n",
       "                           GSM4718581  GSM4718582  GSM4718583  GSM4718584  \\\n",
       "Poor Metabolizer of Drugs           1           1           1           1   \n",
       "\n",
       "                           GSM4718585  GSM4718586  ...  GSM4718659  \\\n",
       "Poor Metabolizer of Drugs           1           1  ...           0   \n",
       "\n",
       "                           GSM4718660  GSM4718661  GSM4718662  GSM4718663  \\\n",
       "Poor Metabolizer of Drugs           0           0           0           0   \n",
       "\n",
       "                           GSM4718664  GSM4718665  GSM4718666  GSM4718667  \\\n",
       "Poor Metabolizer of Drugs           0           0           0           0   \n",
       "\n",
       "                           GSM4718668  \n",
       "Poor Metabolizer of Drugs           0  \n",
       "\n",
       "[1 rows x 62 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM4718577</th>\n",
       "      <th>GSM4718578</th>\n",
       "      <th>GSM4718579</th>\n",
       "      <th>GSM4718580</th>\n",
       "      <th>GSM4718581</th>\n",
       "      <th>GSM4718582</th>\n",
       "      <th>GSM4718583</th>\n",
       "      <th>GSM4718584</th>\n",
       "      <th>GSM4718585</th>\n",
       "      <th>GSM4718586</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4718659</th>\n",
       "      <th>GSM4718660</th>\n",
       "      <th>GSM4718661</th>\n",
       "      <th>GSM4718662</th>\n",
       "      <th>GSM4718663</th>\n",
       "      <th>GSM4718664</th>\n",
       "      <th>GSM4718665</th>\n",
       "      <th>GSM4718666</th>\n",
       "      <th>GSM4718667</th>\n",
       "      <th>GSM4718668</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000029</th>\n",
       "      <td>0.07</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.91</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.11</td>\n",
       "      <td>...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.63</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.69</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000155</th>\n",
       "      <td>0.95</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.96</td>\n",
       "      <td>...</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000158</th>\n",
       "      <td>0.94</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.95</td>\n",
       "      <td>...</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.97</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000165</th>\n",
       "      <td>0.06</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.82</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000221</th>\n",
       "      <td>0.79</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.86</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.78</td>\n",
       "      <td>...</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM4718577  GSM4718578  GSM4718579  GSM4718580  GSM4718581  \\\n",
       "ID                                                                       \n",
       "cg00000029        0.07        0.70        0.29        0.51        0.82   \n",
       "cg00000155        0.95        0.89        0.94        0.90        0.94   \n",
       "cg00000158        0.94        0.92        0.94        0.94        0.95   \n",
       "cg00000165        0.06        0.19        0.05        0.17        0.16   \n",
       "cg00000221        0.79        0.78        0.79        0.80        0.84   \n",
       "\n",
       "            GSM4718582  GSM4718583  GSM4718584  GSM4718585  GSM4718586  ...  \\\n",
       "ID                                                                      ...   \n",
       "cg00000029        0.34        0.52        0.91        0.89        0.11  ...   \n",
       "cg00000155        0.94        0.87        0.90        0.93        0.96  ...   \n",
       "cg00000158        0.96        0.93        0.94        0.94        0.95  ...   \n",
       "cg00000165        0.07        0.06        0.07        0.10        0.08  ...   \n",
       "cg00000221        0.75        0.85        0.86        0.84        0.78  ...   \n",
       "\n",
       "            GSM4718659  GSM4718660  GSM4718661  GSM4718662  GSM4718663  \\\n",
       "ID                                                                       \n",
       "cg00000029        0.46        0.29        0.63        0.54        0.57   \n",
       "cg00000155        0.94        0.92        0.92        0.90        0.94   \n",
       "cg00000158        0.90        0.94        0.89        0.92        0.92   \n",
       "cg00000165        0.59        0.60        0.57        0.72        0.35   \n",
       "cg00000221        0.61        0.76        0.75        0.77        0.81   \n",
       "\n",
       "            GSM4718664  GSM4718665  GSM4718666  GSM4718667  GSM4718668  \n",
       "ID                                                                      \n",
       "cg00000029        0.85        0.69        0.60        0.41        0.69  \n",
       "cg00000155        0.94        0.94        0.97        0.94        0.96  \n",
       "cg00000158        0.96        0.95        0.97        0.89        0.95  \n",
       "cg00000165        0.86        0.82        0.79        0.13        0.50  \n",
       "cg00000221        0.90        0.90        0.94        0.68        0.89  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cg00000029',\n",
       " 'cg00000155',\n",
       " 'cg00000158',\n",
       " 'cg00000165',\n",
       " 'cg00000221',\n",
       " 'cg00000236',\n",
       " 'cg00000292',\n",
       " 'cg00000321',\n",
       " 'cg00000363',\n",
       " 'cg00000540',\n",
       " 'cg00000579',\n",
       " 'cg00000596',\n",
       " 'cg00000622',\n",
       " 'cg00000658',\n",
       " 'cg00000714',\n",
       " 'cg00000721',\n",
       " 'cg00000734',\n",
       " 'cg00000765',\n",
       " 'cg00000769',\n",
       " 'cg00000884']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['cg00000029', 'cg00000155', 'cg00000158', 'cg00000165', 'cg00000221', 'cg00000236', 'cg00000292', 'cg00000321', 'cg00000363', 'cg00000540', 'cg00000579', 'cg00000596', 'cg00000622', 'cg00000658', 'cg00000714', 'cg00000721', 'cg00000734', 'cg00000765', 'cg00000769', 'cg00000884']\\n\""
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'UCSC_RefGene_Name'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['cg00000029', 'cg00000155', 'cg00000158', 'cg00000165', 'cg00000221', 'cg00000236', 'cg00000292', '\n",
      "1000 input query terms found no hit:\t['cg00032376', 'cg00032391', 'cg00032458', 'cg00032544', 'cg00032576', 'cg00032609', 'cg00032610', '\n",
      "1000 input query terms found no hit:\t['cg00066468', 'cg00066499', 'cg00066508', 'cg00066511', 'cg00066535', 'cg00066554', 'cg00066627', '\n",
      "1000 input query terms found no hit:\t['cg00099023', 'cg00099069', 'cg00099091', 'cg00099100', 'cg00099145', 'cg00099216', 'cg00099240', '\n",
      "1000 input query terms found no hit:\t['cg00134295', 'cg00134415', 'cg00134449', 'cg00134504', 'cg00134539', 'cg00134563', 'cg00134602', '\n",
      "1000 input query terms found no hit:\t['cg00169897', 'cg00169917', 'cg00169964', 'cg00170003', 'cg00170056', 'cg00170072', 'cg00170188', '\n",
      "1000 input query terms found no hit:\t['cg00207890', 'cg00207893', 'cg00207921', 'cg00207926', 'cg00207939', 'cg00207947', 'cg00207981', '\n",
      "1000 input query terms found no hit:\t['cg00242949', 'cg00242950', 'cg00242951', 'cg00242965', 'cg00242976', 'cg00243184', 'cg00243187', '\n",
      "1000 input query terms found no hit:\t['cg00280858', 'cg00280870', 'cg00280895', 'cg00280915', 'cg00280922', 'cg00280928', 'cg00280984', '\n",
      "1000 input query terms found no hit:\t['cg00318013', 'cg00318034', 'cg00318059', 'cg00318084', 'cg00318157', 'cg00318242', 'cg00318243', '\n",
      "1000 input query terms found no hit:\t['cg00351011', 'cg00351047', 'cg00351054', 'cg00351079', 'cg00351119', 'cg00351152', 'cg00351190', '\n",
      "1000 input query terms found no hit:\t['cg00386101', 'cg00386141', 'cg00386279', 'cg00386329', 'cg00386338', 'cg00386361', 'cg00386405', '\n",
      "1000 input query terms found no hit:\t['cg00419912', 'cg00419963', 'cg00419990', 'cg00419999', 'cg00420193', 'cg00420212', 'cg00420217', '\n",
      "1000 input query terms found no hit:\t['cg00456162', 'cg00456216', 'cg00456257', 'cg00456299', 'cg00456326', 'cg00456348', 'cg00456370', '\n",
      "1000 input query terms found no hit:\t['cg00489772', 'cg00489795', 'cg00489861', 'cg00489988', 'cg00490052', 'cg00490209', 'cg00490249', '\n",
      "1000 input query terms found no hit:\t['cg00522403', 'cg00522443', 'cg00522450', 'cg00522451', 'cg00522453', 'cg00522555', 'cg00522561', '\n",
      "1000 input query terms found no hit:\t['cg00555420', 'cg00555438', 'cg00555456', 'cg00555538', 'cg00555585', 'cg00555600', 'cg00555675', '\n",
      "1000 input query terms found no hit:\t['cg00589639', 'cg00589699', 'cg00589708', 'cg00589730', 'cg00589760', 'cg00589791', 'cg00589813', '\n",
      "1000 input query terms found no hit:\t['cg00623421', 'cg00623437', 'cg00623462', 'cg00623498', 'cg00623518', 'cg00623593', 'cg00623655', '\n",
      "1000 input query terms found no hit:\t['cg00660106', 'cg00660140', 'cg00660163', 'cg00660167', 'cg00660168', 'cg00660232', 'cg00660251', '\n",
      "1000 input query terms found no hit:\t['cg00696647', 'cg00696671', 'cg00696685', 'cg00696691', 'cg00696756', 'cg00696763', 'cg00696803', '\n",
      "1000 input query terms found no hit:\t['cg00731411', 'cg00731458', 'cg00731485', 'cg00731487', 'cg00731494', 'cg00731526', 'cg00731608', '\n",
      "1000 input query terms found no hit:\t['cg00767429', 'cg00767442', 'cg00767496', 'cg00767503', 'cg00767544', 'cg00767560', 'cg00767573', '\n",
      "1000 input query terms found no hit:\t['cg00803580', 'cg00803638', 'cg00803692', 'cg00803804', 'cg00803816', 'cg00803827', 'cg00803865', '\n",
      "1000 input query terms found no hit:\t['cg00840341', 'cg00840363', 'cg00840370', 'cg00840403', 'cg00840412', 'cg00840433', 'cg00840438', '\n",
      "1000 input query terms found no hit:\t['cg00876462', 'cg00876501', 'cg00876532', 'cg00876541', 'cg00876544', 'cg00876552', 'cg00876571', '\n",
      "1000 input query terms found no hit:\t['cg00913283', 'cg00913288', 'cg00913360', 'cg00913383', 'cg00913391', 'cg00913393', 'cg00913495', '\n",
      "1000 input query terms found no hit:\t['cg00949446', 'cg00949479', 'cg00949511', 'cg00949532', 'cg00949554', 'cg00949578', 'cg00949583', '\n",
      "1000 input query terms found no hit:\t['cg00984905', 'cg00984923', 'cg00984985', 'cg00984997', 'cg00985040', 'cg00985086', 'cg00985104', '\n",
      "1000 input query terms found no hit:\t['cg01020356', 'cg01020368', 'cg01020399', 'cg01020413', 'cg01020421', 'cg01020427', 'cg01020484', '\n",
      "1000 input query terms found no hit:\t['cg01054775', 'cg01054805', 'cg01054832', 'cg01054891', 'cg01054938', 'cg01055000', 'cg01055075', '\n",
      "1000 input query terms found no hit:\t['cg01090256', 'cg01090358', 'cg01090364', 'cg01090433', 'cg01090437', 'cg01090445', 'cg01090508', '\n",
      "1000 input query terms found no hit:\t['cg01125659', 'cg01125676', 'cg01125814', 'cg01125824', 'cg01125846', 'cg01126080', 'cg01126085', '\n",
      "1000 input query terms found no hit:\t['cg01161042', 'cg01161055', 'cg01161087', 'cg01161138', 'cg01161204', 'cg01161216', 'cg01161217', '\n",
      "1000 input query terms found no hit:\t['cg01198208', 'cg01198291', 'cg01198324', 'cg01198338', 'cg01198364', 'cg01198379', 'cg01198391', '\n",
      "1000 input query terms found no hit:\t['cg01235591', 'cg01235596', 'cg01235607', 'cg01235659', 'cg01235742', 'cg01235795', 'cg01235820', '\n",
      "1000 input query terms found no hit:\t['cg01273125', 'cg01273150', 'cg01273194', 'cg01273232', 'cg01273256', 'cg01273264', 'cg01273287', '\n",
      "1000 input query terms found no hit:\t['cg01308536', 'cg01308664', 'cg01308736', 'cg01308737', 'cg01308787', 'cg01308827', 'cg01308832', '\n",
      "1000 input query terms found no hit:\t['cg01346077', 'cg01346114', 'cg01346147', 'cg01346152', 'cg01346158', 'cg01346252', 'cg01346423', '\n",
      "1000 input query terms found no hit:\t['cg01380906', 'cg01380922', 'cg01380946', 'cg01380999', 'cg01381001', 'cg01381109', 'cg01381170', '\n",
      "1000 input query terms found no hit:\t['cg01418409', 'cg01418441', 'cg01418457', 'cg01418527', 'cg01418539', 'cg01418593', 'cg01418618', '\n",
      "1000 input query terms found no hit:\t['cg01453378', 'cg01453386', 'cg01453388', 'cg01453391', 'cg01453441', 'cg01453551', 'cg01453614', '\n",
      "1000 input query terms found no hit:\t['cg01491794', 'cg01491871', 'cg01491910', 'cg01491926', 'cg01491932', 'cg01491990', 'cg01492170', '\n",
      "1000 input query terms found no hit:\t['cg01527261', 'cg01527264', 'cg01527302', 'cg01527307', 'cg01527329', 'cg01527379', 'cg01527394', '\n",
      "1000 input query terms found no hit:\t['cg01563973', 'cg01564033', 'cg01564068', 'cg01564076', 'cg01564135', 'cg01564165', 'cg01564211', '\n",
      "1000 input query terms found no hit:\t['cg01599674', 'cg01599709', 'cg01599714', 'cg01599742', 'cg01599747', 'cg01599770', 'cg01599781', '\n",
      "1000 input query terms found no hit:\t['cg01636300', 'cg01636316', 'cg01636354', 'cg01636393', 'cg01636413', 'cg01636505', 'cg01636541', '\n",
      "1000 input query terms found no hit:\t['cg01676452', 'cg01676467', 'cg01676552', 'cg01676623', 'cg01676680', 'cg01676766', 'cg01676795', '\n",
      "1000 input query terms found no hit:\t['cg01715740', 'cg01715745', 'cg01715763', 'cg01715795', 'cg01715799', 'cg01715800', 'cg01715818', '\n",
      "1000 input query terms found no hit:\t['cg01754162', 'cg01754163', 'cg01754205', 'cg01754219', 'cg01754232', 'cg01754257', 'cg01754267', '\n",
      "1000 input query terms found no hit:\t['cg01790783', 'cg01791000', 'cg01791027', 'cg01791034', 'cg01791062', 'cg01791104', 'cg01791121', '\n",
      "1000 input query terms found no hit:\t['cg01824305', 'cg01824325', 'cg01824349', 'cg01824401', 'cg01824410', 'cg01824412', 'cg01824466', '\n",
      "1000 input query terms found no hit:\t['cg01861913', 'cg01862034', 'cg01862036', 'cg01862052', 'cg01862160', 'cg01862172', 'cg01862248', '\n",
      "1000 input query terms found no hit:\t['cg01896926', 'cg01896999', 'cg01897036', 'cg01897074', 'cg01897092', 'cg01897119', 'cg01897127', '\n",
      "1000 input query terms found no hit:\t['cg01933172', 'cg01933195', 'cg01933235', 'cg01933248', 'cg01933251', 'cg01933276', 'cg01933308', '\n",
      "1000 input query terms found no hit:\t['cg01966465', 'cg01966470', 'cg01966553', 'cg01966561', 'cg01966570', 'cg01966636', 'cg01966674', '\n",
      "1000 input query terms found no hit:\t['cg02004370', 'cg02004381', 'cg02004418', 'cg02004422', 'cg02004437', 'cg02004499', 'cg02004548', '\n",
      "1000 input query terms found no hit:\t['cg02041347', 'cg02041435', 'cg02041470', 'cg02041484', 'cg02041498', 'cg02041547', 'cg02041561', '\n",
      "1000 input query terms found no hit:\t['cg02076518', 'cg02076541', 'cg02076598', 'cg02076604', 'cg02076607', 'cg02076626', 'cg02076629', '\n",
      "1000 input query terms found no hit:\t['cg02111748', 'cg02111792', 'cg02111855', 'cg02111865', 'cg02111944', 'cg02111951', 'cg02111981', '\n",
      "1000 input query terms found no hit:\t['cg02150809', 'cg02150820', 'cg02150867', 'cg02150868', 'cg02150910', 'cg02150912', 'cg02150926', '\n",
      "1000 input query terms found no hit:\t['cg02188190', 'cg02188223', 'cg02188225', 'cg02188299', 'cg02188319', 'cg02188400', 'cg02188443', '\n",
      "1000 input query terms found no hit:\t['cg02226645', 'cg02226672', 'cg02226688', 'cg02226697', 'cg02226779', 'cg02226810', 'cg02226827', '\n",
      "1000 input query terms found no hit:\t['cg02262025', 'cg02262037', 'cg02262056', 'cg02262107', 'cg02262187', 'cg02262195', 'cg02262221', '\n",
      "1000 input query terms found no hit:\t['cg02298523', 'cg02298525', 'cg02298570', 'cg02298574', 'cg02298612', 'cg02298703', 'cg02298721', '\n",
      "1000 input query terms found no hit:\t['cg02332766', 'cg02332769', 'cg02332782', 'cg02332863', 'cg02332902', 'cg02332932', 'cg02332982', '\n",
      "1000 input query terms found no hit:\t['cg02370100', 'cg02370126', 'cg02370222', 'cg02370232', 'cg02370248', 'cg02370334', 'cg02370376', '\n",
      "1000 input query terms found no hit:\t['cg02404346', 'cg02404349', 'cg02404400', 'cg02404410', 'cg02404417', 'cg02404452', 'cg02404484', '\n",
      "1000 input query terms found no hit:\t['cg02445229', 'cg02445282', 'cg02445292', 'cg02445356', 'cg02445421', 'cg02445433', 'cg02445447', '\n",
      "1000 input query terms found no hit:\t['cg02479626', 'cg02479643', 'cg02479656', 'cg02479691', 'cg02479744', 'cg02479755', 'cg02479773', '\n",
      "1000 input query terms found no hit:\t['cg02515468', 'cg02515553', 'cg02515556', 'cg02515559', 'cg02515596', 'cg02515601', 'cg02515646', '\n",
      "1000 input query terms found no hit:\t['cg02551052', 'cg02551095', 'cg02551130', 'cg02551145', 'cg02551218', 'cg02551232', 'cg02551234', '\n",
      "1000 input query terms found no hit:\t['cg02587643', 'cg02587648', 'cg02587685', 'cg02587894', 'cg02587955', 'cg02587988', 'cg02588098', '\n",
      "1000 input query terms found no hit:\t['cg02624051', 'cg02624052', 'cg02624058', 'cg02624067', 'cg02624081', 'cg02624129', 'cg02624246', '\n",
      "1000 input query terms found no hit:\t['cg02657865', 'cg02657889', 'cg02657892', 'cg02657930', 'cg02658046', 'cg02658063', 'cg02658097', '\n",
      "1000 input query terms found no hit:\t['cg02696607', 'cg02696670', 'cg02696675', 'cg02696685', 'cg02696763', 'cg02696790', 'cg02696830', '\n",
      "1000 input query terms found no hit:\t['cg02732358', 'cg02732438', 'cg02732508', 'cg02732509', 'cg02732653', 'cg02732662', 'cg02732671', '\n",
      "1000 input query terms found no hit:\t['cg02767759', 'cg02767771', 'cg02767777', 'cg02767788', 'cg02767827', 'cg02767841', 'cg02767912', '\n",
      "1000 input query terms found no hit:\t['cg02805142', 'cg02805148', 'cg02805188', 'cg02805328', 'cg02805354', 'cg02805356', 'cg02805410', '\n",
      "1000 input query terms found no hit:\t['cg02842711', 'cg02842820', 'cg02842850', 'cg02842869', 'cg02842899', 'cg02842972', 'cg02842991', '\n",
      "1000 input query terms found no hit:\t['cg02878289', 'cg02878355', 'cg02878382', 'cg02878385', 'cg02878414', 'cg02878439', 'cg02878461', '\n",
      "1000 input query terms found no hit:\t['cg02912174', 'cg02912184', 'cg02912187', 'cg02912216', 'cg02912260', 'cg02912279', 'cg02912281', '\n",
      "1000 input query terms found no hit:\t['cg02947862', 'cg02947873', 'cg02947951', 'cg02947993', 'cg02947996', 'cg02947999', 'cg02948067', '\n",
      "1000 input query terms found no hit:\t['cg02981489', 'cg02981620', 'cg02981636', 'cg02981639', 'cg02981663', 'cg02981698', 'cg02981703', '\n",
      "1000 input query terms found no hit:\t['cg03018058', 'cg03018163', 'cg03018256', 'cg03018342', 'cg03018380', 'cg03018478', 'cg03018496', '\n",
      "1000 input query terms found no hit:\t['cg03054157', 'cg03054162', 'cg03054277', 'cg03054288', 'cg03054303', 'cg03054309', 'cg03054343', '\n",
      "1000 input query terms found no hit:\t['cg03091851', 'cg03091982', 'cg03092019', 'cg03092191', 'cg03092252', 'cg03092271', 'cg03092307', '\n",
      "1000 input query terms found no hit:\t['cg03129049', 'cg03129134', 'cg03129172', 'cg03129360', 'cg03129384', 'cg03129394', 'cg03129396', '\n",
      "1000 input query terms found no hit:\t['cg03166265', 'cg03166286', 'cg03166292', 'cg03166302', 'cg03166323', 'cg03166324', 'cg03166482', '\n",
      "1000 input query terms found no hit:\t['cg03203889', 'cg03203942', 'cg03203951', 'cg03204051', 'cg03204109', 'cg03204140', 'cg03204154', '\n",
      "1000 input query terms found no hit:\t['cg03240301', 'cg03240321', 'cg03240324', 'cg03240473', 'cg03240509', 'cg03240518', 'cg03240624', '\n",
      "1000 input query terms found no hit:\t['cg03278568', 'cg03278573', 'cg03278579', 'cg03278639', 'cg03278643', 'cg03278675', 'cg03278715', '\n",
      "1000 input query terms found no hit:\t['cg03315270', 'cg03315344', 'cg03315346', 'cg03315407', 'cg03315432', 'cg03315459', 'cg03315476', '\n",
      "1000 input query terms found no hit:\t['cg03350736', 'cg03350814', 'cg03350828', 'cg03350900', 'cg03350993', 'cg03351003', 'cg03351088', '\n",
      "1000 input query terms found no hit:\t['cg03389364', 'cg03389391', 'cg03389426', 'cg03389538', 'cg03389557', 'cg03389653', 'cg03389662', '\n",
      "1000 input query terms found no hit:\t['cg03426225', 'cg03426226', 'cg03426236', 'cg03426252', 'cg03426320', 'cg03426423', 'cg03426451', '\n",
      "1000 input query terms found no hit:\t['cg03461757', 'cg03461761', 'cg03461777', 'cg03461781', 'cg03461784', 'cg03461851', 'cg03461885', '\n",
      "1000 input query terms found no hit:\t['cg03499581', 'cg03499609', 'cg03499626', 'cg03499639', 'cg03499659', 'cg03499675', 'cg03499775', '\n",
      "1000 input query terms found no hit:\t['cg03536048', 'cg03536100', 'cg03536170', 'cg03536202', 'cg03536229', 'cg03536242', 'cg03536289', '\n",
      "1000 input query terms found no hit:\t['cg03571145', 'cg03571225', 'cg03571290', 'cg03571301', 'cg03571308', 'cg03571320', 'cg03571342', '\n",
      "1000 input query terms found no hit:\t['cg03607825', 'cg03607855', 'cg03607891', 'cg03607907', 'cg03607920', 'cg03607946', 'cg03607972', '\n",
      "1000 input query terms found no hit:\t['cg03644281', 'cg03644378', 'cg03644466', 'cg03644484', 'cg03644514', 'cg03644516', 'cg03644585', '\n",
      "1000 input query terms found no hit:\t['cg03683003', 'cg03683004', 'cg03683031', 'cg03683087', 'cg03683114', 'cg03683132', 'cg03683164', '\n",
      "1000 input query terms found no hit:\t['cg03720687', 'cg03720745', 'cg03720758', 'cg03720762', 'cg03720771', 'cg03720838', 'cg03720842', '\n",
      "1000 input query terms found no hit:\t['cg03755763', 'cg03755794', 'cg03755825', 'cg03755877', 'cg03755888', 'cg03755945', 'cg03755955', '\n",
      "1000 input query terms found no hit:\t['cg03793814', 'cg03793834', 'cg03793846', 'cg03793872', 'cg03793929', 'cg03793937', 'cg03793951', '\n",
      "1000 input query terms found no hit:\t['cg03833329', 'cg03833378', 'cg03833423', 'cg03833427', 'cg03833429', 'cg03833465', 'cg03833466', '\n",
      "1000 input query terms found no hit:\t['cg03872745', 'cg03872783', 'cg03872804', 'cg03872831', 'cg03872846', 'cg03872905', 'cg03872910', '\n",
      "1000 input query terms found no hit:\t['cg03909602', 'cg03909644', 'cg03909721', 'cg03909781', 'cg03909849', 'cg03909863', 'cg03909886', '\n",
      "1000 input query terms found no hit:\t['cg03949033', 'cg03949158', 'cg03949181', 'cg03949274', 'cg03949284', 'cg03949306', 'cg03949362', '\n",
      "1000 input query terms found no hit:\t['cg03986588', 'cg03986620', 'cg03986691', 'cg03986786', 'cg03986812', 'cg03986829', 'cg03986835', '\n",
      "1000 input query terms found no hit:\t['cg04021454', 'cg04021497', 'cg04021502', 'cg04021514', 'cg04021526', 'cg04021549', 'cg04021557', '\n",
      "1000 input query terms found no hit:\t['cg04062391', 'cg04062461', 'cg04062510', 'cg04062529', 'cg04062567', 'cg04062576', 'cg04062622', '\n",
      "1000 input query terms found no hit:\t['cg04099793', 'cg04099803', 'cg04099808', 'cg04099818', 'cg04100007', 'cg04100048', 'cg04100049', '\n",
      "1000 input query terms found no hit:\t['cg04132379', 'cg04132394', 'cg04132407', 'cg04132452', 'cg04132472', 'cg04132498', 'cg04132510', '\n",
      "1000 input query terms found no hit:\t['cg04168185', 'cg04168235', 'cg04168474', 'cg04168479', 'cg04168520', 'cg04168577', 'cg04168609', '\n",
      "1000 input query terms found no hit:\t['cg04205715', 'cg04205744', 'cg04205753', 'cg04205943', 'cg04205950', 'cg04206021', 'cg04206053', '\n",
      "1000 input query terms found no hit:\t['cg04239967', 'cg04239994', 'cg04240077', 'cg04240079', 'cg04240109', 'cg04240200', 'cg04240267', '\n",
      "1000 input query terms found no hit:\t['cg04276245', 'cg04276288', 'cg04276301', 'cg04276302', 'cg04276316', 'cg04276370', 'cg04276413', '\n",
      "1000 input query terms found no hit:\t['cg04314071', 'cg04314111', 'cg04314114', 'cg04314130', 'cg04314154', 'cg04314155', 'cg04314218', '\n",
      "1000 input query terms found no hit:\t['cg04350645', 'cg04350646', 'cg04350675', 'cg04350772', 'cg04350922', 'cg04351018', 'cg04351046', '\n",
      "1000 input query terms found no hit:\t['cg04389822', 'cg04389838', 'cg04389853', 'cg04389885', 'cg04389899', 'cg04389926', 'cg04389950', '\n",
      "1000 input query terms found no hit:\t['cg04425687', 'cg04425701', 'cg04425710', 'cg04425783', 'cg04425787', 'cg04425788', 'cg04425820', '\n",
      "1000 input query terms found no hit:\t['cg04465350', 'cg04465425', 'cg04465473', 'cg04465474', 'cg04465525', 'cg04465599', 'cg04465730', '\n",
      "1000 input query terms found no hit:\t['cg04503054', 'cg04503063', 'cg04503093', 'cg04503160', 'cg04503171', 'cg04503240', 'cg04503267', '\n",
      "1000 input query terms found no hit:\t['cg04539895', 'cg04539934', 'cg04539935', 'cg04539944', 'cg04539957', 'cg04540039', 'cg04540061', '\n",
      "1000 input query terms found no hit:\t['cg04574459', 'cg04574480', 'cg04574507', 'cg04574532', 'cg04574700', 'cg04574716', 'cg04574732', '\n",
      "1000 input query terms found no hit:\t['cg04610242', 'cg04610274', 'cg04610325', 'cg04610375', 'cg04610410', 'cg04610422', 'cg04610432', '\n",
      "1000 input query terms found no hit:\t['cg04651914', 'cg04652066', 'cg04652097', 'cg04652257', 'cg04652270', 'cg04652282', 'cg04652333', '\n",
      "1000 input query terms found no hit:\t['cg04689058', 'cg04689061', 'cg04689080', 'cg04689082', 'cg04689131', 'cg04689178', 'cg04689331', '\n",
      "1000 input query terms found no hit:\t['cg04728310', 'cg04728329', 'cg04728363', 'cg04728402', 'cg04728429', 'cg04728446', 'cg04728480', '\n",
      "1000 input query terms found no hit:\t['cg04766076', 'cg04766084', 'cg04766136', 'cg04766178', 'cg04766198', 'cg04766250', 'cg04766289', '\n",
      "1000 input query terms found no hit:\t['cg04802018', 'cg04802034', 'cg04802040', 'cg04802171', 'cg04802221', 'cg04802236', 'cg04802238', '\n",
      "1000 input query terms found no hit:\t['cg04839710', 'cg04839725', 'cg04839741', 'cg04839761', 'cg04839764', 'cg04839779', 'cg04839780', '\n",
      "1000 input query terms found no hit:\t['cg04874358', 'cg04874375', 'cg04874446', 'cg04874463', 'cg04874511', 'cg04874530', 'cg04874562', '\n",
      "1000 input query terms found no hit:\t['cg04908655', 'cg04908668', 'cg04908782', 'cg04908789', 'cg04908905', 'cg04908907', 'cg04908911', '\n",
      "1000 input query terms found no hit:\t['cg04945434', 'cg04945443', 'cg04945457', 'cg04945511', 'cg04945512', 'cg04945541', 'cg04945542', '\n",
      "1000 input query terms found no hit:\t['cg04983142', 'cg04983146', 'cg04983148', 'cg04983163', 'cg04983247', 'cg04983254', 'cg04983276', '\n",
      "1000 input query terms found no hit:\t['cg05018638', 'cg05018648', 'cg05018671', 'cg05018681', 'cg05018741', 'cg05018821', 'cg05018835', '\n",
      "1000 input query terms found no hit:\t['cg05054898', 'cg05054904', 'cg05054910', 'cg05054953', 'cg05054998', 'cg05055073', 'cg05055108', '\n",
      "1000 input query terms found no hit:\t['cg05091734', 'cg05091736', 'cg05091796', 'cg05091831', 'cg05091838', 'cg05091919', 'cg05091920', '\n",
      "1000 input query terms found no hit:\t['cg05131191', 'cg05131196', 'cg05131202', 'cg05131207', 'cg05131254', 'cg05131266', 'cg05131334', '\n",
      "1000 input query terms found no hit:\t['cg05166744', 'cg05166801', 'cg05166819', 'cg05166820', 'cg05166851', 'cg05166871', 'cg05166976', '\n",
      "1000 input query terms found no hit:\t['cg05204618', 'cg05204638', 'cg05204673', 'cg05204754', 'cg05204798', 'cg05204809', 'cg05204830', '\n",
      "1000 input query terms found no hit:\t['cg05239790', 'cg05239791', 'cg05239799', 'cg05239811', 'cg05239841', 'cg05239872', 'cg05239909', '\n",
      "1000 input query terms found no hit:\t['cg05276155', 'cg05276175', 'cg05276182', 'cg05276184', 'cg05276195', 'cg05276226', 'cg05276243', '\n",
      "1000 input query terms found no hit:\t['cg05313595', 'cg05313644', 'cg05313647', 'cg05313656', 'cg05313680', 'cg05313743', 'cg05313771', '\n",
      "1000 input query terms found no hit:\t['cg05349788', 'cg05349837', 'cg05349845', 'cg05349900', 'cg05349919', 'cg05349922', 'cg05349952', '\n",
      "1000 input query terms found no hit:\t['cg05383910', 'cg05383931', 'cg05383947', 'cg05383958', 'cg05384053', 'cg05384075', 'cg05384123', '\n",
      "1000 input query terms found no hit:\t['cg05420867', 'cg05420899', 'cg05420975', 'cg05420994', 'cg05421036', 'cg05421040', 'cg05421097', '\n",
      "1000 input query terms found no hit:\t['cg05462446', 'cg05462557', 'cg05462570', 'cg05462578', 'cg05462625', 'cg05462639', 'cg05462658', '\n",
      "1000 input query terms found no hit:\t['cg05500106', 'cg05500125', 'cg05500240', 'cg05500341', 'cg05500400', 'cg05500413', 'cg05500549', '\n",
      "1000 input query terms found no hit:\t['cg05539991', 'cg05540047', 'cg05540052', 'cg05540062', 'cg05540100', 'cg05540133', 'cg05540221', '\n",
      "1000 input query terms found no hit:\t['cg05579534', 'cg05579549', 'cg05579559', 'cg05579573', 'cg05579575', 'cg05579580', 'cg05579598', '\n",
      "1000 input query terms found no hit:\t['cg05617452', 'cg05617469', 'cg05617483', 'cg05617551', 'cg05617623', 'cg05617650', 'cg05617732', '\n",
      "1000 input query terms found no hit:\t['cg05653385', 'cg05653400', 'cg05653472', 'cg05653482', 'cg05653496', 'cg05653535', 'cg05653627', '\n",
      "1000 input query terms found no hit:\t['cg05692130', 'cg05692188', 'cg05692292', 'cg05692319', 'cg05692420', 'cg05692440', 'cg05692532', '\n",
      "1000 input query terms found no hit:\t['cg05729118', 'cg05729185', 'cg05729199', 'cg05729200', 'cg05729249', 'cg05729309', 'cg05729352', '\n",
      "1000 input query terms found no hit:\t['cg05768367', 'cg05768403', 'cg05768419', 'cg05768427', 'cg05768444', 'cg05768532', 'cg05768558', '\n",
      "1000 input query terms found no hit:\t['cg05806233', 'cg05806255', 'cg05806312', 'cg05806333', 'cg05806375', 'cg05806434', 'cg05806455', '\n",
      "1000 input query terms found no hit:\t['cg05847163', 'cg05847183', 'cg05847191', 'cg05847200', 'cg05847233', 'cg05847388', 'cg05847519', '\n",
      "1000 input query terms found no hit:\t['cg05883874', 'cg05883907', 'cg05883932', 'cg05883951', 'cg05883978', 'cg05884023', 'cg05884032', '\n",
      "1000 input query terms found no hit:\t['cg05919376', 'cg05919454', 'cg05919460', 'cg05919461', 'cg05919507', 'cg05919523', 'cg05919561', '\n",
      "1000 input query terms found no hit:\t['cg05953296', 'cg05953340', 'cg05953364', 'cg05953373', 'cg05953394', 'cg05953477', 'cg05953549', '\n",
      "1000 input query terms found no hit:\t['cg05991220', 'cg05991223', 'cg05991314', 'cg05991331', 'cg05991333', 'cg05991342', 'cg05991388', '\n",
      "1000 input query terms found no hit:\t['cg06030274', 'cg06030290', 'cg06030310', 'cg06030362', 'cg06030377', 'cg06030456', 'cg06030459', '\n",
      "1000 input query terms found no hit:\t['cg06069429', 'cg06069457', 'cg06069504', 'cg06069616', 'cg06069740', 'cg06069861', 'cg06069877', '\n",
      "1000 input query terms found no hit:\t['cg06106756', 'cg06106763', 'cg06106793', 'cg06106810', 'cg06106839', 'cg06106883', 'cg06106905', '\n",
      "1000 input query terms found no hit:\t['cg06146795', 'cg06146802', 'cg06146897', 'cg06146910', 'cg06146930', 'cg06146943', 'cg06146964', '\n",
      "1000 input query terms found no hit:\t['cg06185924', 'cg06185934', 'cg06185965', 'cg06186041', 'cg06186057', 'cg06186108', 'cg06186120', '\n",
      "1000 input query terms found no hit:\t['cg06222220', 'cg06222383', 'cg06222392', 'cg06222397', 'cg06222414', 'cg06222419', 'cg06222469', '\n",
      "1000 input query terms found no hit:\t['cg06263372', 'cg06263438', 'cg06263461', 'cg06263473', 'cg06263495', 'cg06263630', 'cg06263672', '\n",
      "1000 input query terms found no hit:\t['cg06304070', 'cg06304074', 'cg06304140', 'cg06304167', 'cg06304190', 'cg06304205', 'cg06304316', '\n",
      "1000 input query terms found no hit:\t['cg06343615', 'cg06343651', 'cg06343669', 'cg06343673', 'cg06343732', 'cg06343739', 'cg06343740', '\n",
      "1000 input query terms found no hit:\t['cg06381693', 'cg06381725', 'cg06381772', 'cg06381803', 'cg06381831', 'cg06381846', 'cg06381879', '\n",
      "1000 input query terms found no hit:\t['cg06419959', 'cg06419964', 'cg06420019', 'cg06420033', 'cg06420049', 'cg06420072', 'cg06420074', '\n",
      "1000 input query terms found no hit:\t['cg06460328', 'cg06460400', 'cg06460511', 'cg06460553', 'cg06460568', 'cg06460573', 'cg06460600', '\n",
      "1000 input query terms found no hit:\t['cg06500017', 'cg06500042', 'cg06500079', 'cg06500090', 'cg06500096', 'cg06500120', 'cg06500134', '\n",
      "1000 input query terms found no hit:\t['cg06536150', 'cg06536167', 'cg06536200', 'cg06536252', 'cg06536308', 'cg06536333', 'cg06536469', '\n",
      "1000 input query terms found no hit:\t['cg06573143', 'cg06573146', 'cg06573181', 'cg06573245', 'cg06573254', 'cg06573295', 'cg06573331', '\n",
      "1000 input query terms found no hit:\t['cg06610988', 'cg06611075', 'cg06611076', 'cg06611115', 'cg06611143', 'cg06611277', 'cg06611310', '\n",
      "1000 input query terms found no hit:\t['cg06646588', 'cg06646604', 'cg06646622', 'cg06646636', 'cg06646646', 'cg06646650', 'cg06646679', '\n",
      "1000 input query terms found no hit:\t['cg06682219', 'cg06682252', 'cg06682282', 'cg06682286', 'cg06682330', 'cg06682369', 'cg06682371', '\n",
      "1000 input query terms found no hit:\t['cg06717615', 'cg06717633', 'cg06717650', 'cg06717725', 'cg06717747', 'cg06717750', 'cg06717754', '\n",
      "1000 input query terms found no hit:\t['cg06753273', 'cg06753303', 'cg06753316', 'cg06753321', 'cg06753333', 'cg06753367', 'cg06753398', '\n",
      "1000 input query terms found no hit:\t['cg06788565', 'cg06788631', 'cg06788675', 'cg06788697', 'cg06788751', 'cg06788767', 'cg06788790', '\n",
      "1000 input query terms found no hit:\t['cg06824308', 'cg06824318', 'cg06824332', 'cg06824394', 'cg06824412', 'cg06824423', 'cg06824443', '\n",
      "1000 input query terms found no hit:\t['cg06861266', 'cg06861301', 'cg06861330', 'cg06861343', 'cg06861370', 'cg06861375', 'cg06861426', '\n",
      "1000 input query terms found no hit:\t['cg06894462', 'cg06894464', 'cg06894469', 'cg06894549', 'cg06894585', 'cg06894599', 'cg06894600', '\n",
      "1000 input query terms found no hit:\t['cg06929521', 'cg06929583', 'cg06929595', 'cg06929622', 'cg06929627', 'cg06929662', 'cg06929724', '\n",
      "1000 input query terms found no hit:\t['cg06965961', 'cg06966028', 'cg06966082', 'cg06966092', 'cg06966110', 'cg06966113', 'cg06966178', '\n",
      "1000 input query terms found no hit:\t['cg07003587', 'cg07003611', 'cg07003617', 'cg07003632', 'cg07003643', 'cg07003645', 'cg07003668', '\n",
      "1000 input query terms found no hit:\t['cg07038921', 'cg07038973', 'cg07038977', 'cg07039074', 'cg07039130', 'cg07039172', 'cg07039180', '\n",
      "1000 input query terms found no hit:\t['cg07076525', 'cg07076661', 'cg07076711', 'cg07076751', 'cg07076792', 'cg07076816', 'cg07076836', '\n",
      "1000 input query terms found no hit:\t['cg07112997', 'cg07113006', 'cg07113025', 'cg07113062', 'cg07113065', 'cg07113119', 'cg07113136', '\n",
      "1000 input query terms found no hit:\t['cg07148914', 'cg07148919', 'cg07149011', 'cg07149017', 'cg07149022', 'cg07149030', 'cg07149033', '\n",
      "1000 input query terms found no hit:\t['cg07183148', 'cg07183218', 'cg07183229', 'cg07183242', 'cg07183316', 'cg07183320', 'cg07183344', '\n",
      "1000 input query terms found no hit:\t['cg07221309', 'cg07221438', 'cg07221444', 'cg07221454', 'cg07221496', 'cg07221511', 'cg07221518', '\n",
      "1000 input query terms found no hit:\t['cg07259541', 'cg07259581', 'cg07259651', 'cg07259660', 'cg07259683', 'cg07259710', 'cg07259711', '\n",
      "1000 input query terms found no hit:\t['cg07297540', 'cg07297582', 'cg07297604', 'cg07297702', 'cg07297764', 'cg07297777', 'cg07297802', '\n",
      "1000 input query terms found no hit:\t['cg07333040', 'cg07333046', 'cg07333061', 'cg07333129', 'cg07333191', 'cg07333223', 'cg07333254', '\n",
      "1000 input query terms found no hit:\t['cg07370900', 'cg07370905', 'cg07370959', 'cg07370966', 'cg07370974', 'cg07370994', 'cg07371029', '\n",
      "1000 input query terms found no hit:\t['cg07409897', 'cg07410037', 'cg07410044', 'cg07410075', 'cg07410086', 'cg07410108', 'cg07410176', '\n",
      "1000 input query terms found no hit:\t['cg07450213', 'cg07450214', 'cg07450219', 'cg07450275', 'cg07450292', 'cg07450298', 'cg07450341', '\n",
      "1000 input query terms found no hit:\t['cg07485496', 'cg07485518', 'cg07485529', 'cg07485541', 'cg07485554', 'cg07485685', 'cg07485723', '\n",
      "1000 input query terms found no hit:\t['cg07523904', 'cg07523958', 'cg07524000', 'cg07524053', 'cg07524057', 'cg07524137', 'cg07524214', '\n",
      "1000 input query terms found no hit:\t['cg07561894', 'cg07562002', 'cg07562079', 'cg07562120', 'cg07562135', 'cg07562175', 'cg07562195', '\n",
      "1000 input query terms found no hit:\t['cg07598324', 'cg07598331', 'cg07598341', 'cg07598357', 'cg07598367', 'cg07598385', 'cg07598407', '\n",
      "1000 input query terms found no hit:\t['cg07635485', 'cg07635530', 'cg07635567', 'cg07635579', 'cg07635586', 'cg07635623', 'cg07635625', '\n",
      "1000 input query terms found no hit:\t['cg07674897', 'cg07674974', 'cg07675031', 'cg07675080', 'cg07675122', 'cg07675221', 'cg07675259', '\n",
      "1000 input query terms found no hit:\t['cg07713027', 'cg07713031', 'cg07713066', 'cg07713120', 'cg07713135', 'cg07713194', 'cg07713268', '\n",
      "1000 input query terms found no hit:\t['cg07749243', 'cg07749266', 'cg07749294', 'cg07749353', 'cg07749412', 'cg07749442', 'cg07749454', '\n",
      "1000 input query terms found no hit:\t['cg07790935', 'cg07790947', 'cg07790960', 'cg07791014', 'cg07791017', 'cg07791039', 'cg07791058', '\n",
      "1000 input query terms found no hit:\t['cg07827181', 'cg07827280', 'cg07827281', 'cg07827304', 'cg07827337', 'cg07827347', 'cg07827395', '\n",
      "1000 input query terms found no hit:\t['cg07866137', 'cg07866179', 'cg07866206', 'cg07866207', 'cg07866212', 'cg07866246', 'cg07866293', '\n",
      "1000 input query terms found no hit:\t['cg07904817', 'cg07904834', 'cg07904845', 'cg07904865', 'cg07904879', 'cg07904900', 'cg07904962', '\n",
      "1000 input query terms found no hit:\t['cg07941301', 'cg07941396', 'cg07941411', 'cg07941474', 'cg07941529', 'cg07941542', 'cg07941567', '\n",
      "1000 input query terms found no hit:\t['cg07977933', 'cg07977946', 'cg07977990', 'cg07978005', 'cg07978029', 'cg07978099', 'cg07978145', '\n",
      "1000 input query terms found no hit:\t['cg08018049', 'cg08018179', 'cg08018188', 'cg08018256', 'cg08018303', 'cg08018356', 'cg08018468', '\n",
      "1000 input query terms found no hit:\t['cg08056738', 'cg08056755', 'cg08056778', 'cg08056831', 'cg08056835', 'cg08056850', 'cg08056863', '\n",
      "1000 input query terms found no hit:\t['cg08093619', 'cg08093719', 'cg08093733', 'cg08093892', 'cg08093928', 'cg08093959', 'cg08093983', '\n",
      "1000 input query terms found no hit:\t['cg08129420', 'cg08129449', 'cg08129471', 'cg08129475', 'cg08129505', 'cg08129583', 'cg08129673', '\n",
      "1000 input query terms found no hit:\t['cg08168435', 'cg08168529', 'cg08168580', 'cg08168585', 'cg08168630', 'cg08168669', 'cg08168689', '\n",
      "1000 input query terms found no hit:\t['cg08206308', 'cg08206318', 'cg08206357', 'cg08206517', 'cg08206536', 'cg08206537', 'cg08206623', '\n",
      "1000 input query terms found no hit:\t['cg08243728', 'cg08243778', 'cg08243782', 'cg08243790', 'cg08243827', 'cg08243849', 'cg08243909', '\n",
      "1000 input query terms found no hit:\t['cg08279131', 'cg08279171', 'cg08279184', 'cg08279186', 'cg08279213', 'cg08279214', 'cg08279217', '\n",
      "1000 input query terms found no hit:\t['cg08317267', 'cg08317373', 'cg08317380', 'cg08317412', 'cg08317416', 'cg08317460', 'cg08317493', '\n",
      "1000 input query terms found no hit:\t['cg08358793', 'cg08358800', 'cg08358803', 'cg08358813', 'cg08358816', 'cg08358898', 'cg08358907', '\n",
      "1000 input query terms found no hit:\t['cg08398452', 'cg08398473', 'cg08398554', 'cg08398556', 'cg08398567', 'cg08398596', 'cg08398691', '\n",
      "1000 input query terms found no hit:\t['cg08439109', 'cg08439120', 'cg08439122', 'cg08439149', 'cg08439216', 'cg08439244', 'cg08439271', '\n",
      "1000 input query terms found no hit:\t['cg08474669', 'cg08474716', 'cg08474748', 'cg08474775', 'cg08474786', 'cg08474813', 'cg08474826', '\n",
      "1000 input query terms found no hit:\t['cg08512791', 'cg08512806', 'cg08512837', 'cg08512847', 'cg08512849', 'cg08512876', 'cg08512882', '\n",
      "1000 input query terms found no hit:\t['cg08548444', 'cg08548445', 'cg08548478', 'cg08548486', 'cg08548489', 'cg08548500', 'cg08548535', '\n",
      "1000 input query terms found no hit:\t['cg08586669', 'cg08586692', 'cg08586737', 'cg08586809', 'cg08586826', 'cg08586849', 'cg08586855', '\n",
      "1000 input query terms found no hit:\t['cg08626833', 'cg08626876', 'cg08626932', 'cg08626939', 'cg08627010', 'cg08627047', 'cg08627051', '\n",
      "1000 input query terms found no hit:\t['cg08665076', 'cg08665079', 'cg08665162', 'cg08665193', 'cg08665196', 'cg08665249', 'cg08665251', '\n",
      "1000 input query terms found no hit:\t['cg08701998', 'cg08702001', 'cg08702007', 'cg08702052', 'cg08702160', 'cg08702225', 'cg08702249', '\n",
      "1000 input query terms found no hit:\t['cg08740223', 'cg08740227', 'cg08740240', 'cg08740284', 'cg08740365', 'cg08740424', 'cg08740509', '\n",
      "1000 input query terms found no hit:\t['cg08780527', 'cg08780538', 'cg08780545', 'cg08780600', 'cg08780617', 'cg08780680', 'cg08780684', '\n",
      "1000 input query terms found no hit:\t['cg08818337', 'cg08818377', 'cg08818385', 'cg08818525', 'cg08818610', 'cg08818632', 'cg08818782', '\n",
      "1000 input query terms found no hit:\t['cg08854511', 'cg08854560', 'cg08854582', 'cg08854628', 'cg08854706', 'cg08854732', 'cg08854791', '\n",
      "1000 input query terms found no hit:\t['cg08887933', 'cg08888038', 'cg08888043', 'cg08888053', 'cg08888071', 'cg08888170', 'cg08888178', '\n",
      "1000 input query terms found no hit:\t['cg08924969', 'cg08924995', 'cg08925019', 'cg08925046', 'cg08925102', 'cg08925129', 'cg08925142', '\n",
      "1000 input query terms found no hit:\t['cg08963716', 'cg08963753', 'cg08963876', 'cg08963966', 'cg08963972', 'cg08964024', 'cg08964099', '\n",
      "1000 input query terms found no hit:\t['cg09005115', 'cg09005125', 'cg09005159', 'cg09005171', 'cg09005221', 'cg09005230', 'cg09005378', '\n",
      "1000 input query terms found no hit:\t['cg09043294', 'cg09043302', 'cg09043403', 'cg09043456', 'cg09043486', 'cg09043511', 'cg09043518', '\n",
      "1000 input query terms found no hit:\t['cg09085355', 'cg09085358', 'cg09085417', 'cg09085428', 'cg09085439', 'cg09085501', 'cg09085507', '\n",
      "1000 input query terms found no hit:\t['cg09124532', 'cg09124558', 'cg09124602', 'cg09124636', 'cg09124679', 'cg09124703', 'cg09124749', '\n",
      "1000 input query terms found no hit:\t['cg09161455', 'cg09161513', 'cg09161528', 'cg09161542', 'cg09161549', 'cg09161570', 'cg09161579', '\n",
      "1000 input query terms found no hit:\t['cg09201130', 'cg09201151', 'cg09201221', 'cg09201238', 'cg09201264', 'cg09201266', 'cg09201327', '\n",
      "1000 input query terms found no hit:\t['cg09238743', 'cg09238746', 'cg09238756', 'cg09238801', 'cg09238886', 'cg09238890', 'cg09238899', '\n",
      "1000 input query terms found no hit:\t['cg09275704', 'cg09275729', 'cg09275735', 'cg09275783', 'cg09275862', 'cg09275869', 'cg09275910', '\n",
      "1000 input query terms found no hit:\t['cg09312809', 'cg09312814', 'cg09312895', 'cg09312904', 'cg09312906', 'cg09312914', 'cg09312947', '\n",
      "1000 input query terms found no hit:\t['cg09350131', 'cg09350274', 'cg09350341', 'cg09350362', 'cg09350387', 'cg09350411', 'cg09350414', '\n",
      "1000 input query terms found no hit:\t['cg09388810', 'cg09388860', 'cg09388868', 'cg09388920', 'cg09388991', 'cg09389069', 'cg09389147', '\n",
      "1000 input query terms found no hit:\t['cg09427077', 'cg09427114', 'cg09427137', 'cg09427156', 'cg09427208', 'cg09427273', 'cg09427311', '\n",
      "1000 input query terms found no hit:\t['cg09465065', 'cg09465142', 'cg09465149', 'cg09465168', 'cg09465239', 'cg09465288', 'cg09465326', '\n",
      "1000 input query terms found no hit:\t['cg09504498', 'cg09504571', 'cg09504585', 'cg09504612', 'cg09504658', 'cg09504692', 'cg09504737', '\n",
      "1000 input query terms found no hit:\t['cg09543660', 'cg09543693', 'cg09543727', 'cg09543792', 'cg09543819', 'cg09543842', 'cg09543853', '\n",
      "1000 input query terms found no hit:\t['cg09581197', 'cg09581216', 'cg09581226', 'cg09581281', 'cg09581413', 'cg09581452', 'cg09581513', '\n",
      "1000 input query terms found no hit:\t['cg09618971', 'cg09618988', 'cg09618998', 'cg09619020', 'cg09619064', 'cg09619146', 'cg09619271', '\n",
      "1000 input query terms found no hit:\t['cg09657538', 'cg09657542', 'cg09657667', 'cg09657673', 'cg09657712', 'cg09657744', 'cg09657852', '\n",
      "1000 input query terms found no hit:\t['cg09697033', 'cg09697084', 'cg09697104', 'cg09697245', 'cg09697259', 'cg09697386', 'cg09697403', '\n",
      "1000 input query terms found no hit:\t['cg09738074', 'cg09738100', 'cg09738143', 'cg09738156', 'cg09738190', 'cg09738193', 'cg09738214', '\n",
      "1000 input query terms found no hit:\t['cg09777883', 'cg09778187', 'cg09778229', 'cg09778245', 'cg09778271', 'cg09778357', 'cg09778401', '\n",
      "1000 input query terms found no hit:\t['cg09817374', 'cg09817427', 'cg09817431', 'cg09817455', 'cg09817503', 'cg09817523', 'cg09817738', '\n",
      "1000 input query terms found no hit:\t['cg09854775', 'cg09854828', 'cg09854852', 'cg09855000', 'cg09855001', 'cg09855037', 'cg09855089', '\n",
      "1000 input query terms found no hit:\t['cg09890775', 'cg09890800', 'cg09890870', 'cg09890887', 'cg09890891', 'cg09890892', 'cg09890963', '\n",
      "1000 input query terms found no hit:\t['cg09930712', 'cg09930713', 'cg09930729', 'cg09930743', 'cg09930748', 'cg09930772', 'cg09930908', '\n",
      "1000 input query terms found no hit:\t['cg09969827', 'cg09969829', 'cg09969830', 'cg09969919', 'cg09969989', 'cg09970021', 'cg09970023', '\n",
      "1000 input query terms found no hit:\t['cg10008625', 'cg10008636', 'cg10008710', 'cg10008725', 'cg10008731', 'cg10008733', 'cg10008820', '\n",
      "1000 input query terms found no hit:\t['cg10046227', 'cg10046270', 'cg10046367', 'cg10046388', 'cg10046451', 'cg10046616', 'cg10046620', '\n",
      "1000 input query terms found no hit:\t['cg10085623', 'cg10085651', 'cg10085668', 'cg10085741', 'cg10085750', 'cg10085824', 'cg10085844', '\n",
      "1000 input query terms found no hit:\t['cg10122097', 'cg10122116', 'cg10122156', 'cg10122231', 'cg10122245', 'cg10122252', 'cg10122267', '\n",
      "1000 input query terms found no hit:\t['cg10158841', 'cg10158843', 'cg10158850', 'cg10158886', 'cg10158975', 'cg10158997', 'cg10159006', '\n",
      "1000 input query terms found no hit:\t['cg10194844', 'cg10194870', 'cg10194884', 'cg10194942', 'cg10195011', 'cg10195050', 'cg10195146', '\n",
      "1000 input query terms found no hit:\t['cg10231737', 'cg10231739', 'cg10231742', 'cg10231746', 'cg10231785', 'cg10231801', 'cg10231862', '\n",
      "1000 input query terms found no hit:\t['cg10270719', 'cg10270745', 'cg10270770', 'cg10270798', 'cg10270839', 'cg10270866', 'cg10271003', '\n",
      "1000 input query terms found no hit:\t['cg10310081', 'cg10310111', 'cg10310119', 'cg10310141', 'cg10310144', 'cg10310179', 'cg10310183', '\n",
      "1000 input query terms found no hit:\t['cg10347426', 'cg10347456', 'cg10347474', 'cg10347660', 'cg10347735', 'cg10347759', 'cg10347799', '\n",
      "1000 input query terms found no hit:\t['cg10386887', 'cg10386914', 'cg10386924', 'cg10387007', 'cg10387195', 'cg10387280', 'cg10387289', '\n",
      "1000 input query terms found no hit:\t['cg10428661', 'cg10428791', 'cg10428805', 'cg10428850', 'cg10428879', 'cg10428881', 'cg10428982', '\n",
      "1000 input query terms found no hit:\t['cg10467015', 'cg10467021', 'cg10467022', 'cg10467038', 'cg10467073', 'cg10467201', 'cg10467217', '\n",
      "1000 input query terms found no hit:\t['cg10503827', 'cg10503854', 'cg10503867', 'cg10503891', 'cg10503992', 'cg10504000', 'cg10504003', '\n",
      "1000 input query terms found no hit:\t['cg10539616', 'cg10539631', 'cg10539700', 'cg10539808', 'cg10539928', 'cg10539956', 'cg10540110', '\n",
      "1000 input query terms found no hit:\t['cg10580691', 'cg10580715', 'cg10580753', 'cg10580786', 'cg10580816', 'cg10580832', 'cg10580903', '\n",
      "1000 input query terms found no hit:\t['cg10617363', 'cg10617419', 'cg10617494', 'cg10617499', 'cg10617577', 'cg10617605', 'cg10617617', '\n",
      "1000 input query terms found no hit:\t['cg10658197', 'cg10658203', 'cg10658252', 'cg10658277', 'cg10658307', 'cg10658349', 'cg10658358', '\n",
      "1000 input query terms found no hit:\t['cg10697938', 'cg10697992', 'cg10697995', 'cg10698097', 'cg10698098', 'cg10698155', 'cg10698161', '\n",
      "1000 input query terms found no hit:\t['cg10736770', 'cg10736786', 'cg10736791', 'cg10736843', 'cg10736902', 'cg10736906', 'cg10736914', '\n",
      "1000 input query terms found no hit:\t['cg10777220', 'cg10777276', 'cg10777347', 'cg10777361', 'cg10777388', 'cg10777461', 'cg10777462', '\n",
      "1000 input query terms found no hit:\t['cg10815470', 'cg10815483', 'cg10815486', 'cg10815543', 'cg10815625', 'cg10815723', 'cg10815745', '\n",
      "1000 input query terms found no hit:\t['cg10854110', 'cg10854158', 'cg10854181', 'cg10854209', 'cg10854292', 'cg10854423', 'cg10854441', '\n",
      "1000 input query terms found no hit:\t['cg10896683', 'cg10896746', 'cg10896765', 'cg10896774', 'cg10896796', 'cg10896810', 'cg10896862', '\n",
      "1000 input query terms found no hit:\t['cg10937302', 'cg10937341', 'cg10937372', 'cg10937408', 'cg10937428', 'cg10937494', 'cg10937511', '\n",
      "1000 input query terms found no hit:\t['cg10977352', 'cg10977398', 'cg10977428', 'cg10977501', 'cg10977619', 'cg10977643', 'cg10977648', '\n",
      "1000 input query terms found no hit:\t['cg11017613', 'cg11017765', 'cg11017923', 'cg11017935', 'cg11017981', 'cg11018013', 'cg11018018', '\n",
      "1000 input query terms found no hit:\t['cg11057824', 'cg11057867', 'cg11057877', 'cg11057897', 'cg11057900', 'cg11057934', 'cg11058037', '\n",
      "1000 input query terms found no hit:\t['cg11099496', 'cg11099555', 'cg11099586', 'cg11099598', 'cg11099651', 'cg11099655', 'cg11099683', '\n",
      "1000 input query terms found no hit:\t['cg11140393', 'cg11140440', 'cg11140591', 'cg11140595', 'cg11140606', 'cg11140679', 'cg11140740', '\n",
      "1000 input query terms found no hit:\t['cg11181726', 'cg11181750', 'cg11181763', 'cg11181795', 'cg11181817', 'cg11181985', 'cg11182018', '\n",
      "1000 input query terms found no hit:\t['cg11218915', 'cg11218974', 'cg11218979', 'cg11219022', 'cg11219030', 'cg11219066', 'cg11219097', '\n",
      "1000 input query terms found no hit:\t['cg11256374', 'cg11256387', 'cg11256445', 'cg11256458', 'cg11256548', 'cg11256607', 'cg11256628', '\n",
      "1000 input query terms found no hit:\t['cg11296945', 'cg11296946', 'cg11297013', 'cg11297016', 'cg11297046', 'cg11297107', 'cg11297127', '\n",
      "1000 input query terms found no hit:\t['cg11336000', 'cg11336053', 'cg11336061', 'cg11336159', 'cg11336178', 'cg11336206', 'cg11336232', '\n",
      "1000 input query terms found no hit:\t['cg11374345', 'cg11374355', 'cg11374368', 'cg11374425', 'cg11374432', 'cg11374434', 'cg11374463', '\n",
      "1000 input query terms found no hit:\t['cg11413039', 'cg11413098', 'cg11413127', 'cg11413133', 'cg11413443', 'cg11413498', 'cg11413506', '\n",
      "1000 input query terms found no hit:\t['cg11453530', 'cg11453585', 'cg11453603', 'cg11453608', 'cg11453712', 'cg11453719', 'cg11453771', '\n",
      "1000 input query terms found no hit:\t['cg11496792', 'cg11496800', 'cg11496811', 'cg11497131', 'cg11497147', 'cg11497335', 'cg11497372', '\n",
      "1000 input query terms found no hit:\t['cg11539171', 'cg11539173', 'cg11539188', 'cg11539275', 'cg11539296', 'cg11539302', 'cg11539351', '\n",
      "1000 input query terms found no hit:\t['cg11580575', 'cg11580596', 'cg11580663', 'cg11580676', 'cg11580702', 'cg11580704', 'cg11580759', '\n",
      "1000 input query terms found no hit:\t['cg11621437', 'cg11621523', 'cg11621559', 'cg11621586', 'cg11621608', 'cg11621647', 'cg11621667', '\n",
      "1000 input query terms found no hit:\t['cg11659961', 'cg11660018', 'cg11660071', 'cg11660146', 'cg11660304', 'cg11660314', 'cg11660352', '\n",
      "1000 input query terms found no hit:\t['cg11699258', 'cg11699261', 'cg11699334', 'cg11699374', 'cg11699403', 'cg11699435', 'cg11699436', '\n",
      "1000 input query terms found no hit:\t['cg11737193', 'cg11737230', 'cg11737232', 'cg11737246', 'cg11737251', 'cg11737298', 'cg11737334', '\n",
      "1000 input query terms found no hit:\t['cg11776344', 'cg11776349', 'cg11776350', 'cg11776520', 'cg11776534', 'cg11776562', 'cg11776586', '\n",
      "1000 input query terms found no hit:\t['cg11813940', 'cg11813976', 'cg11814024', 'cg11814059', 'cg11814087', 'cg11814155', 'cg11814211', '\n",
      "1000 input query terms found no hit:\t['cg11850639', 'cg11850773', 'cg11850802', 'cg11850853', 'cg11850861', 'cg11850961', 'cg11851098', '\n",
      "1000 input query terms found no hit:\t['cg11889145', 'cg11889164', 'cg11889195', 'cg11889265', 'cg11889334', 'cg11889342', 'cg11889371', '\n",
      "1000 input query terms found no hit:\t['cg11931123', 'cg11931139', 'cg11931203', 'cg11931211', 'cg11931220', 'cg11931223', 'cg11931235', '\n",
      "1000 input query terms found no hit:\t['cg11968713', 'cg11968742', 'cg11968792', 'cg11968804', 'cg11968877', 'cg11968890', 'cg11968932', '\n",
      "1000 input query terms found no hit:\t['cg12009803', 'cg12009816', 'cg12009872', 'cg12009882', 'cg12009969', 'cg12009994', 'cg12010071', '\n",
      "1000 input query terms found no hit:\t['cg12049872', 'cg12049875', 'cg12049897', 'cg12049925', 'cg12049949', 'cg12049961', 'cg12049973', '\n",
      "1000 input query terms found no hit:\t['cg12085660', 'cg12085672', 'cg12085698', 'cg12085812', 'cg12085842', 'cg12085846', 'cg12085903', '\n",
      "1000 input query terms found no hit:\t['cg12126035', 'cg12126037', 'cg12126038', 'cg12126043', 'cg12126243', 'cg12126298', 'cg12126344', '\n",
      "1000 input query terms found no hit:\t['cg12165223', 'cg12165225', 'cg12165236', 'cg12165240', 'cg12165250', 'cg12165265', 'cg12165471', '\n",
      "1000 input query terms found no hit:\t['cg12205193', 'cg12205206', 'cg12205230', 'cg12205241', 'cg12205244', 'cg12205270', 'cg12205358', '\n",
      "1000 input query terms found no hit:\t['cg12245084', 'cg12245130', 'cg12245188', 'cg12245206', 'cg12245236', 'cg12245245', 'cg12245247', '\n",
      "1000 input query terms found no hit:\t['cg12287702', 'cg12287785', 'cg12287833', 'cg12287936', 'cg12287945', 'cg12287976', 'cg12287987', '\n",
      "1000 input query terms found no hit:\t['cg12330441', 'cg12330703', 'cg12330709', 'cg12330748', 'cg12330854', 'cg12330998', 'cg12331004', '\n",
      "1000 input query terms found no hit:\t['cg12373920', 'cg12373934', 'cg12373951', 'cg12373979', 'cg12373982', 'cg12374003', 'cg12374021', '\n",
      "1000 input query terms found no hit:\t['cg12414526', 'cg12414557', 'cg12414653', 'cg12414681', 'cg12414731', 'cg12414746', 'cg12414759', '\n",
      "1000 input query terms found no hit:\t['cg12451530', 'cg12451605', 'cg12451606', 'cg12451671', 'cg12451676', 'cg12451679', 'cg12451731', '\n",
      "1000 input query terms found no hit:\t['cg12489934', 'cg12489936', 'cg12489956', 'cg12489960', 'cg12490092', 'cg12490365', 'cg12490395', '\n",
      "1000 input query terms found no hit:\t['cg12530491', 'cg12530497', 'cg12530503', 'cg12530572', 'cg12530665', 'cg12530740', 'cg12531003', '\n",
      "1000 input query terms found no hit:\t['cg12571007', 'cg12571055', 'cg12571072', 'cg12571098', 'cg12571155', 'cg12571241', 'cg12571300', '\n",
      "1000 input query terms found no hit:\t['cg12608775', 'cg12608784', 'cg12608881', 'cg12608898', 'cg12609030', 'cg12609052', 'cg12609063', '\n",
      "1000 input query terms found no hit:\t['cg12648213', 'cg12648261', 'cg12648263', 'cg12648297', 'cg12648336', 'cg12648392', 'cg12648510', '\n",
      "1000 input query terms found no hit:\t['cg12687767', 'cg12687813', 'cg12687824', 'cg12687886', 'cg12687897', 'cg12687922', 'cg12687926', '\n",
      "1000 input query terms found no hit:\t['cg12730702', 'cg12730771', 'cg12730802', 'cg12730826', 'cg12730832', 'cg12730853', 'cg12731010', '\n",
      "1000 input query terms found no hit:\t['cg12771165', 'cg12771168', 'cg12771177', 'cg12771178', 'cg12771255', 'cg12771257', 'cg12771281', '\n",
      "1000 input query terms found no hit:\t['cg12811648', 'cg12811653', 'cg12811753', 'cg12811795', 'cg12811797', 'cg12811871', 'cg12811901', '\n",
      "1000 input query terms found no hit:\t['cg12853083', 'cg12853145', 'cg12853178', 'cg12853184', 'cg12853199', 'cg12853247', 'cg12853277', '\n",
      "1000 input query terms found no hit:\t['cg12894427', 'cg12894449', 'cg12894478', 'cg12894524', 'cg12894525', 'cg12894629', 'cg12894649', '\n",
      "1000 input query terms found no hit:\t['cg12935350', 'cg12935359', 'cg12935434', 'cg12935455', 'cg12935497', 'cg12935627', 'cg12935694', '\n",
      "1000 input query terms found no hit:\t['cg12975224', 'cg12975230', 'cg12975324', 'cg12975330', 'cg12975390', 'cg12975399', 'cg12975461', '\n",
      "1000 input query terms found no hit:\t['cg13017339', 'cg13017345', 'cg13017464', 'cg13017471', 'cg13017478', 'cg13017573', 'cg13017603', '\n",
      "1000 input query terms found no hit:\t['cg13057632', 'cg13057681', 'cg13057731', 'cg13057808', 'cg13057840', 'cg13057857', 'cg13057875', '\n",
      "1000 input query terms found no hit:\t['cg13097886', 'cg13097987', 'cg13097993', 'cg13098071', 'cg13098082', 'cg13098091', 'cg13098164', '\n",
      "1000 input query terms found no hit:\t['cg13141115', 'cg13141188', 'cg13141192', 'cg13141224', 'cg13141246', 'cg13141281', 'cg13141289', '\n",
      "1000 input query terms found no hit:\t['cg13181026', 'cg13181052', 'cg13181079', 'cg13181164', 'cg13181251', 'cg13181287', 'cg13181291', '\n",
      "1000 input query terms found no hit:\t['cg13223473', 'cg13223509', 'cg13223510', 'cg13223537', 'cg13223565', 'cg13223652', 'cg13223677', '\n",
      "1000 input query terms found no hit:\t['cg13264115', 'cg13264143', 'cg13264159', 'cg13264208', 'cg13264209', 'cg13264217', 'cg13264234', '\n",
      "1000 input query terms found no hit:\t['cg13302587', 'cg13302642', 'cg13302664', 'cg13302670', 'cg13302767', 'cg13302823', 'cg13302832', '\n",
      "1000 input query terms found no hit:\t['cg13343368', 'cg13343392', 'cg13343486', 'cg13343502', 'cg13343549', 'cg13343565', 'cg13343638', '\n",
      "1000 input query terms found no hit:\t['cg13380601', 'cg13380610', 'cg13380612', 'cg13380622', 'cg13380624', 'cg13380643', 'cg13380669', '\n",
      "1000 input query terms found no hit:\t['cg13418196', 'cg13418221', 'cg13418230', 'cg13418283', 'cg13418303', 'cg13418419', 'cg13418456', '\n",
      "1000 input query terms found no hit:\t['cg13453157', 'cg13453162', 'cg13453168', 'cg13453175', 'cg13453193', 'cg13453196', 'cg13453203', '\n",
      "1000 input query terms found no hit:\t['cg13487505', 'cg13487523', 'cg13487553', 'cg13487575', 'cg13487631', 'cg13487642', 'cg13487666', '\n",
      "1000 input query terms found no hit:\t['cg13523713', 'cg13523731', 'cg13523771', 'cg13523818', 'cg13523819', 'cg13523875', 'cg13523882', '\n",
      "1000 input query terms found no hit:\t['cg13558568', 'cg13558608', 'cg13558627', 'cg13558665', 'cg13558682', 'cg13558754', 'cg13558764', '\n",
      "1000 input query terms found no hit:\t['cg13592941', 'cg13592980', 'cg13592982', 'cg13593009', 'cg13593090', 'cg13593092', 'cg13593132', '\n",
      "1000 input query terms found no hit:\t['cg13627467', 'cg13627479', 'cg13627491', 'cg13627498', 'cg13627562', 'cg13627599', 'cg13627659', '\n",
      "1000 input query terms found no hit:\t['cg13664067', 'cg13664146', 'cg13664178', 'cg13664195', 'cg13664201', 'cg13664233', 'cg13664345', '\n",
      "1000 input query terms found no hit:\t['cg13698574', 'cg13698578', 'cg13698611', 'cg13698613', 'cg13698691', 'cg13698714', 'cg13698778', '\n",
      "1000 input query terms found no hit:\t['cg13734658', 'cg13734661', 'cg13734707', 'cg13734718', 'cg13734729', 'cg13734780', 'cg13734792', '\n",
      "1000 input query terms found no hit:\t['cg13773898', 'cg13773914', 'cg13773948', 'cg13773959', 'cg13773968', 'cg13774005', 'cg13774015', '\n",
      "1000 input query terms found no hit:\t['cg13808083', 'cg13808088', 'cg13808090', 'cg13808141', 'cg13808145', 'cg13808147', 'cg13808160', '\n",
      "1000 input query terms found no hit:\t['cg13845634', 'cg13845656', 'cg13845692', 'cg13845792', 'cg13845800', 'cg13845858', 'cg13845952', '\n",
      "1000 input query terms found no hit:\t['cg13882837', 'cg13882918', 'cg13882921', 'cg13882988', 'cg13883063', 'cg13883113', 'cg13883127', '\n",
      "1000 input query terms found no hit:\t['cg13916990', 'cg13917004', 'cg13917020', 'cg13917023', 'cg13917047', 'cg13917060', 'cg13917099', '\n",
      "1000 input query terms found no hit:\t['cg13952106', 'cg13952159', 'cg13952242', 'cg13952315', 'cg13952398', 'cg13952447', 'cg13952483', '\n",
      "1000 input query terms found no hit:\t['cg13987792', 'cg13987794', 'cg13987862', 'cg13987879', 'cg13987885', 'cg13987911', 'cg13987922', '\n",
      "1000 input query terms found no hit:\t['cg14022183', 'cg14022202', 'cg14022258', 'cg14022313', 'cg14022322', 'cg14022358', 'cg14022377', '\n",
      "1000 input query terms found no hit:\t['cg14055957', 'cg14055970', 'cg14055996', 'cg14056065', 'cg14056110', 'cg14056206', 'cg14056226', '\n",
      "1000 input query terms found no hit:\t['cg14093018', 'cg14093065', 'cg14093101', 'cg14093102', 'cg14093103', 'cg14093106', 'cg14093125', '\n",
      "1000 input query terms found no hit:\t['cg14128773', 'cg14128812', 'cg14128819', 'cg14128890', 'cg14128911', 'cg14128915', 'cg14128925', '\n",
      "1000 input query terms found no hit:\t['cg14164380', 'cg14164492', 'cg14164556', 'cg14164691', 'cg14164816', 'cg14164839', 'cg14164880', '\n",
      "1000 input query terms found no hit:\t['cg14200594', 'cg14200609', 'cg14200631', 'cg14200633', 'cg14200635', 'cg14200692', 'cg14200725', '\n",
      "1000 input query terms found no hit:\t['cg14233374', 'cg14233375', 'cg14233448', 'cg14233507', 'cg14233565', 'cg14233568', 'cg14233578', '\n",
      "1000 input query terms found no hit:\t['cg14268155', 'cg14268172', 'cg14268184', 'cg14268200', 'cg14268223', 'cg14268226', 'cg14268271', '\n",
      "1000 input query terms found no hit:\t['cg14303437', 'cg14303464', 'cg14303478', 'cg14303481', 'cg14303526', 'cg14303674', 'cg14303699', '\n",
      "1000 input query terms found no hit:\t['cg14338226', 'cg14338234', 'cg14338246', 'cg14338256', 'cg14338299', 'cg14338425', 'cg14338472', '\n",
      "1000 input query terms found no hit:\t['cg14375955', 'cg14376031', 'cg14376033', 'cg14376041', 'cg14376106', 'cg14376170', 'cg14376275', '\n",
      "1000 input query terms found no hit:\t['cg14413209', 'cg14413262', 'cg14413263', 'cg14413285', 'cg14413354', 'cg14413368', 'cg14413378', '\n",
      "1000 input query terms found no hit:\t['cg14446575', 'cg14446588', 'cg14446604', 'cg14446613', 'cg14446622', 'cg14446658', 'cg14446682', '\n",
      "1000 input query terms found no hit:\t['cg14481528', 'cg14481604', 'cg14481613', 'cg14481619', 'cg14481689', 'cg14481832', 'cg14481911', '\n",
      "1000 input query terms found no hit:\t['cg14517901', 'cg14517917', 'cg14517940', 'cg14517946', 'cg14517983', 'cg14518039', 'cg14518091', '\n",
      "1000 input query terms found no hit:\t['cg14550457', 'cg14550518', 'cg14550539', 'cg14550559', 'cg14550563', 'cg14550570', 'cg14550667', '\n",
      "1000 input query terms found no hit:\t['cg14584529', 'cg14584535', 'cg14584577', 'cg14584616', 'cg14584636', 'cg14584702', 'cg14584716', '\n",
      "1000 input query terms found no hit:\t['cg14621053', 'cg14621090', 'cg14621103', 'cg14621146', 'cg14621209', 'cg14621217', 'cg14621221', '\n",
      "1000 input query terms found no hit:\t['cg14657710', 'cg14657725', 'cg14657728', 'cg14657818', 'cg14657834', 'cg14657876', 'cg14657915', '\n",
      "1000 input query terms found no hit:\t['cg14695540', 'cg14695566', 'cg14695609', 'cg14695662', 'cg14695663', 'cg14695751', 'cg14695804', '\n",
      "1000 input query terms found no hit:\t['cg14737131', 'cg14737169', 'cg14737211', 'cg14737251', 'cg14737257', 'cg14737263', 'cg14737286', '\n",
      "1000 input query terms found no hit:\t['cg14777143', 'cg14777209', 'cg14777226', 'cg14777244', 'cg14777258', 'cg14777295', 'cg14777341', '\n",
      "1000 input query terms found no hit:\t['cg14817241', 'cg14817346', 'cg14817370', 'cg14817440', 'cg14817448', 'cg14817487', 'cg14817501', '\n",
      "1000 input query terms found no hit:\t['cg14852134', 'cg14852227', 'cg14852238', 'cg14852267', 'cg14852276', 'cg14852310', 'cg14852355', '\n",
      "1000 input query terms found no hit:\t['cg14891553', 'cg14891611', 'cg14891617', 'cg14891651', 'cg14891733', 'cg14891734', 'cg14891762', '\n",
      "1000 input query terms found no hit:\t['cg14932016', 'cg14932109', 'cg14932124', 'cg14932133', 'cg14932200', 'cg14932219', 'cg14932237', '\n",
      "1000 input query terms found no hit:\t['cg14974371', 'cg14974383', 'cg14974384', 'cg14974397', 'cg14974411', 'cg14974448', 'cg14974572', '\n",
      "1000 input query terms found no hit:\t['cg15014008', 'cg15014010', 'cg15014037', 'cg15014038', 'cg15014069', 'cg15014116', 'cg15014127', '\n",
      "1000 input query terms found no hit:\t['cg15054873', 'cg15054954', 'cg15054972', 'cg15054977', 'cg15055039', 'cg15055060', 'cg15055095', '\n",
      "1000 input query terms found no hit:\t['cg15091323', 'cg15091333', 'cg15091337', 'cg15091407', 'cg15091419', 'cg15091424', 'cg15091508', '\n",
      "1000 input query terms found no hit:\t['cg15130252', 'cg15130253', 'cg15130302', 'cg15130342', 'cg15130384', 'cg15130433', 'cg15130454', '\n",
      "1000 input query terms found no hit:\t['cg15168727', 'cg15168787', 'cg15168798', 'cg15168816', 'cg15168825', 'cg15168842', 'cg15168850', '\n",
      "1000 input query terms found no hit:\t['cg15212585', 'cg15212590', 'cg15212712', 'cg15212814', 'cg15212821', 'cg15212895', 'cg15212899', '\n",
      "1000 input query terms found no hit:\t['cg15254064', 'cg15254147', 'cg15254238', 'cg15254287', 'cg15254385', 'cg15254390', 'cg15254424', '\n",
      "1000 input query terms found no hit:\t['cg15298722', 'cg15298791', 'cg15298793', 'cg15298796', 'cg15298831', 'cg15298833', 'cg15298875', '\n",
      "1000 input query terms found no hit:\t['cg15340652', 'cg15340705', 'cg15340727', 'cg15340776', 'cg15340845', 'cg15340874', 'cg15340886', '\n",
      "1000 input query terms found no hit:\t['cg15383827', 'cg15383847', 'cg15383879', 'cg15383900', 'cg15383962', 'cg15383979', 'cg15383980', '\n",
      "1000 input query terms found no hit:\t['cg15423407', 'cg15423452', 'cg15423496', 'cg15423581', 'cg15423587', 'cg15423613', 'cg15423654', '\n",
      "1000 input query terms found no hit:\t['cg15463628', 'cg15463636', 'cg15463664', 'cg15463686', 'cg15463692', 'cg15463714', 'cg15463769', '\n",
      "1000 input query terms found no hit:\t['cg15506477', 'cg15506486', 'cg15506514', 'cg15506515', 'cg15506577', 'cg15506587', 'cg15506591', '\n",
      "1000 input query terms found no hit:\t['cg15550572', 'cg15550589', 'cg15550618', 'cg15550741', 'cg15550769', 'cg15550804', 'cg15550930', '\n",
      "1000 input query terms found no hit:\t['cg15590426', 'cg15590461', 'cg15590502', 'cg15590606', 'cg15590671', 'cg15590692', 'cg15590712', '\n",
      "1000 input query terms found no hit:\t['cg15629912', 'cg15629948', 'cg15630071', 'cg15630089', 'cg15630215', 'cg15630265', 'cg15630287', '\n",
      "1000 input query terms found no hit:\t['cg15669314', 'cg15669357', 'cg15669600', 'cg15669605', 'cg15669692', 'cg15669749', 'cg15669803', '\n",
      "1000 input query terms found no hit:\t['cg15707553', 'cg15707579', 'cg15707705', 'cg15707747', 'cg15707857', 'cg15707869', 'cg15707948', '\n",
      "1000 input query terms found no hit:\t['cg15745265', 'cg15745273', 'cg15745284', 'cg15745385', 'cg15745450', 'cg15745469', 'cg15745560', '\n",
      "1000 input query terms found no hit:\t['cg15789069', 'cg15789095', 'cg15789106', 'cg15789107', 'cg15789150', 'cg15789211', 'cg15789260', '\n",
      "1000 input query terms found no hit:\t['cg15826006', 'cg15826083', 'cg15826130', 'cg15826273', 'cg15826352', 'cg15826390', 'cg15826398', '\n",
      "1000 input query terms found no hit:\t['cg15863924', 'cg15863932', 'cg15864003', 'cg15864101', 'cg15864104', 'cg15864134', 'cg15864140', '\n",
      "1000 input query terms found no hit:\t['cg15903819', 'cg15903837', 'cg15903860', 'cg15903872', 'cg15903883', 'cg15903887', 'cg15903915', '\n",
      "1000 input query terms found no hit:\t['cg15942535', 'cg15942562', 'cg15942569', 'cg15942601', 'cg15942624', 'cg15942641', 'cg15942666', '\n",
      "1000 input query terms found no hit:\t['cg15985184', 'cg15985191', 'cg15985196', 'cg15985218', 'cg15985280', 'cg15985297', 'cg15985344', '\n",
      "1000 input query terms found no hit:\t['cg16020986', 'cg16021052', 'cg16021113', 'cg16021119', 'cg16021126', 'cg16021135', 'cg16021157', '\n",
      "1000 input query terms found no hit:\t['cg16061253', 'cg16061254', 'cg16061301', 'cg16061354', 'cg16061375', 'cg16061381', 'cg16061399', '\n",
      "1000 input query terms found no hit:\t['cg16102240', 'cg16102247', 'cg16102251', 'cg16102262', 'cg16102283', 'cg16102291', 'cg16102333', '\n",
      "1000 input query terms found no hit:\t['cg16143491', 'cg16143502', 'cg16143506', 'cg16143519', 'cg16143569', 'cg16143578', 'cg16143626', '\n",
      "1000 input query terms found no hit:\t['cg16182658', 'cg16182691', 'cg16182699', 'cg16182707', 'cg16182713', 'cg16182743', 'cg16182860', '\n",
      "1000 input query terms found no hit:\t['cg16223158', 'cg16223190', 'cg16223220', 'cg16223358', 'cg16223436', 'cg16223541', 'cg16223546', '\n",
      "1000 input query terms found no hit:\t['cg16260720', 'cg16260758', 'cg16260768', 'cg16260807', 'cg16260889', 'cg16260931', 'cg16260976', '\n",
      "1000 input query terms found no hit:\t['cg16299455', 'cg16299517', 'cg16299524', 'cg16299537', 'cg16299559', 'cg16299565', 'cg16299576', '\n",
      "1000 input query terms found no hit:\t['cg16336506', 'cg16336556', 'cg16336586', 'cg16336621', 'cg16336651', 'cg16336655', 'cg16336665', '\n",
      "1000 input query terms found no hit:\t['cg16375728', 'cg16375742', 'cg16375820', 'cg16375824', 'cg16375880', 'cg16375909', 'cg16375948', '\n",
      "1000 input query terms found no hit:\t['cg16412973', 'cg16412986', 'cg16412995', 'cg16413020', 'cg16413023', 'cg16413026', 'cg16413043', '\n",
      "1000 input query terms found no hit:\t['cg16450290', 'cg16450309', 'cg16450330', 'cg16450432', 'cg16450547', 'cg16450654', 'cg16450683', '\n",
      "1000 input query terms found no hit:\t['cg16491668', 'cg16491677', 'cg16491682', 'cg16491694', 'cg16491729', 'cg16491734', 'cg16491739', '\n",
      "1000 input query terms found no hit:\t['cg16529268', 'cg16529318', 'cg16529440', 'cg16529472', 'cg16529477', 'cg16529520', 'cg16529548', '\n",
      "1000 input query terms found no hit:\t['cg16566737', 'cg16566825', 'cg16566899', 'cg16566911', 'cg16566943', 'cg16566986', 'cg16567044', '\n",
      "1000 input query terms found no hit:\t['cg16604567', 'cg16604658', 'cg16604680', 'cg16604801', 'cg16604812', 'cg16604835', 'cg16604837', '\n",
      "1000 input query terms found no hit:\t['cg16644143', 'cg16644176', 'cg16644177', 'cg16644209', 'cg16644245', 'cg16644301', 'cg16644329', '\n",
      "1000 input query terms found no hit:\t['cg16679650', 'cg16679651', 'cg16679708', 'cg16679797', 'cg16679832', 'cg16680018', 'cg16680125', '\n",
      "1000 input query terms found no hit:\t['cg16715717', 'cg16715727', 'cg16715785', 'cg16715831', 'cg16715883', 'cg16715937', 'cg16715953', '\n",
      "1000 input query terms found no hit:\t['cg16755239', 'cg16755251', 'cg16755348', 'cg16755393', 'cg16755428', 'cg16755475', 'cg16755494', '\n",
      "1000 input query terms found no hit:\t['cg16795266', 'cg16795280', 'cg16795290', 'cg16795295', 'cg16795304', 'cg16795307', 'cg16795352', '\n",
      "1000 input query terms found no hit:\t['cg16838259', 'cg16838301', 'cg16838358', 'cg16838359', 'cg16838503', 'cg16838536', 'cg16838554', '\n",
      "1000 input query terms found no hit:\t['cg16881891', 'cg16881960', 'cg16882054', 'cg16882097', 'cg16882206', 'cg16882226', 'cg16882301', '\n",
      "1000 input query terms found no hit:\t['cg16920303', 'cg16920407', 'cg16920446', 'cg16920487', 'cg16920502', 'cg16920538', 'cg16920551', '\n",
      "1000 input query terms found no hit:\t['cg16961070', 'cg16961080', 'cg16961092', 'cg16961111', 'cg16961218', 'cg16961235', 'cg16961295', '\n",
      "1000 input query terms found no hit:\t['cg17002899', 'cg17002920', 'cg17002993', 'cg17002994', 'cg17003018', 'cg17003143', 'cg17003199', '\n",
      "1000 input query terms found no hit:\t['cg17043823', 'cg17043828', 'cg17043864', 'cg17043875', 'cg17043919', 'cg17043957', 'cg17043988', '\n",
      "1000 input query terms found no hit:\t['cg17083281', 'cg17083323', 'cg17083342', 'cg17083390', 'cg17083411', 'cg17083429', 'cg17083469', '\n",
      "1000 input query terms found no hit:\t['cg17123676', 'cg17123690', 'cg17123718', 'cg17123728', 'cg17123795', 'cg17123809', 'cg17123837', '\n",
      "1000 input query terms found no hit:\t['cg17161921', 'cg17161924', 'cg17161933', 'cg17162024', 'cg17162028', 'cg17162031', 'cg17162069', '\n",
      "1000 input query terms found no hit:\t['cg17201192', 'cg17201225', 'cg17201227', 'cg17201273', 'cg17201307', 'cg17201343', 'cg17201438', '\n",
      "1000 input query terms found no hit:\t['cg17241935', 'cg17241943', 'cg17241948', 'cg17241999', 'cg17242143', 'cg17242144', 'cg17242241', '\n",
      "1000 input query terms found no hit:\t['cg17283101', 'cg17283111', 'cg17283117', 'cg17283137', 'cg17283139', 'cg17283169', 'cg17283196', '\n",
      "1000 input query terms found no hit:\t['cg17321214', 'cg17321235', 'cg17321267', 'cg17321296', 'cg17321298', 'cg17321329', 'cg17321346', '\n",
      "1000 input query terms found no hit:\t['cg17357132', 'cg17357149', 'cg17357161', 'cg17357214', 'cg17357243', 'cg17357247', 'cg17357258', '\n",
      "1000 input query terms found no hit:\t['cg17398251', 'cg17398255', 'cg17398261', 'cg17398329', 'cg17398388', 'cg17398517', 'cg17398538', '\n",
      "1000 input query terms found no hit:\t['cg17437852', 'cg17437868', 'cg17437930', 'cg17437939', 'cg17437946', 'cg17438030', 'cg17438055', '\n",
      "1000 input query terms found no hit:\t['cg17476389', 'cg17476405', 'cg17476421', 'cg17476499', 'cg17476518', 'cg17476649', 'cg17476701', '\n",
      "1000 input query terms found no hit:\t['cg17516138', 'cg17516156', 'cg17516221', 'cg17516242', 'cg17516247', 'cg17516271', 'cg17516330', '\n",
      "1000 input query terms found no hit:\t['cg17556601', 'cg17556612', 'cg17556788', 'cg17556833', 'cg17556976', 'cg17557063', 'cg17557099', '\n",
      "1000 input query terms found no hit:\t['cg17600358', 'cg17600393', 'cg17600402', 'cg17600444', 'cg17600630', 'cg17600659', 'cg17600661', '\n",
      "1000 input query terms found no hit:\t['cg17643388', 'cg17643402', 'cg17643446', 'cg17643517', 'cg17643564', 'cg17643598', 'cg17643624', '\n",
      "1000 input query terms found no hit:\t['cg17685579', 'cg17685622', 'cg17685628', 'cg17685630', 'cg17685679', 'cg17685724', 'cg17685758', '\n",
      "1000 input query terms found no hit:\t['cg17728974', 'cg17728987', 'cg17729019', 'cg17729249', 'cg17729253', 'cg17729301', 'cg17729322', '\n",
      "1000 input query terms found no hit:\t['cg17770792', 'cg17770844', 'cg17770886', 'cg17770888', 'cg17770910', 'cg17770994', 'cg17771011', '\n",
      "1000 input query terms found no hit:\t['cg17814717', 'cg17814753', 'cg17814758', 'cg17814781', 'cg17814814', 'cg17814977', 'cg17815035', '\n",
      "1000 input query terms found no hit:\t['cg17856610', 'cg17856611', 'cg17856763', 'cg17856830', 'cg17856907', 'cg17856915', 'cg17856981', '\n",
      "1000 input query terms found no hit:\t['cg17899343', 'cg17899373', 'cg17899417', 'cg17899428', 'cg17899476', 'cg17899492', 'cg17899606', '\n",
      "1000 input query terms found no hit:\t['cg17946161', 'cg17946216', 'cg17946266', 'cg17946279', 'cg17946284', 'cg17946316', 'cg17946375', '\n",
      "1000 input query terms found no hit:\t['cg17991843', 'cg17991940', 'cg17991988', 'cg17991990', 'cg17991991', 'cg17992018', 'cg17992034', '\n",
      "1000 input query terms found no hit:\t['cg18035016', 'cg18035056', 'cg18035136', 'cg18035158', 'cg18035229', 'cg18035255', 'cg18035291', '\n",
      "1000 input query terms found no hit:\t['cg18080211', 'cg18080256', 'cg18080303', 'cg18080370', 'cg18080401', 'cg18080500', 'cg18080509', '\n",
      "1000 input query terms found no hit:\t['cg18120049', 'cg18120053', 'cg18120111', 'cg18120193', 'cg18120259', 'cg18120278', 'cg18120303', '\n",
      "1000 input query terms found no hit:\t['cg18162758', 'cg18162783', 'cg18162875', 'cg18162948', 'cg18163062', 'cg18163075', 'cg18163092', '\n",
      "1000 input query terms found no hit:\t['cg18205205', 'cg18205372', 'cg18205376', 'cg18205413', 'cg18205516', 'cg18205517', 'cg18205595', '\n",
      "1000 input query terms found no hit:\t['cg18248284', 'cg18248290', 'cg18248375', 'cg18248420', 'cg18248493', 'cg18248499', 'cg18248586', '\n",
      "1000 input query terms found no hit:\t['cg18294483', 'cg18294502', 'cg18294563', 'cg18294631', 'cg18294646', 'cg18294652', 'cg18294691', '\n",
      "1000 input query terms found no hit:\t['cg18339118', 'cg18339298', 'cg18339303', 'cg18339364', 'cg18339422', 'cg18339493', 'cg18339519', '\n",
      "1000 input query terms found no hit:\t['cg18384976', 'cg18385050', 'cg18385125', 'cg18385190', 'cg18385233', 'cg18385254', 'cg18385299', '\n",
      "1000 input query terms found no hit:\t['cg18430991', 'cg18431002', 'cg18431100', 'cg18431119', 'cg18431123', 'cg18431127', 'cg18431149', '\n",
      "1000 input query terms found no hit:\t['cg18472694', 'cg18472787', 'cg18472881', 'cg18472912', 'cg18472945', 'cg18473026', 'cg18473033', '\n",
      "1000 input query terms found no hit:\t['cg18517266', 'cg18517369', 'cg18517474', 'cg18517517', 'cg18517534', 'cg18517540', 'cg18517542', '\n",
      "1000 input query terms found no hit:\t['cg18562525', 'cg18562570', 'cg18562663', 'cg18562689', 'cg18562714', 'cg18562896', 'cg18562923', '\n",
      "1000 input query terms found no hit:\t['cg18605377', 'cg18605382', 'cg18605392', 'cg18605441', 'cg18605491', 'cg18605534', 'cg18605585', '\n",
      "1000 input query terms found no hit:\t['cg18651663', 'cg18651727', 'cg18651790', 'cg18651821', 'cg18651840', 'cg18651842', 'cg18651864', '\n",
      "1000 input query terms found no hit:\t['cg18698654', 'cg18698681', 'cg18698690', 'cg18698699', 'cg18698767', 'cg18698788', 'cg18698819', '\n",
      "1000 input query terms found no hit:\t['cg18741810', 'cg18741833', 'cg18741908', 'cg18741951', 'cg18741958', 'cg18741962', 'cg18741999', '\n",
      "1000 input query terms found no hit:\t['cg18784565', 'cg18784574', 'cg18784586', 'cg18784592', 'cg18784831', 'cg18784903', 'cg18784912', '\n",
      "1000 input query terms found no hit:\t['cg18826637', 'cg18826684', 'cg18826718', 'cg18826771', 'cg18826787', 'cg18826875', 'cg18826899', '\n",
      "1000 input query terms found no hit:\t['cg18871757', 'cg18871807', 'cg18871861', 'cg18871892', 'cg18871984', 'cg18871988', 'cg18872016', '\n",
      "1000 input query terms found no hit:\t['cg18918538', 'cg18918547', 'cg18918588', 'cg18918774', 'cg18918831', 'cg18918891', 'cg18919017', '\n",
      "1000 input query terms found no hit:\t['cg18967846', 'cg18968018', 'cg18968090', 'cg18968241', 'cg18968256', 'cg18968279', 'cg18968305', '\n",
      "1000 input query terms found no hit:\t['cg19014705', 'cg19014730', 'cg19014749', 'cg19014792', 'cg19014858', 'cg19014941', 'cg19015032', '\n",
      "1000 input query terms found no hit:\t['cg19059348', 'cg19059413', 'cg19059423', 'cg19059495', 'cg19059694', 'cg19059861', 'cg19059954', '\n",
      "1000 input query terms found no hit:\t['cg19106087', 'cg19106154', 'cg19106168', 'cg19106195', 'cg19106237', 'cg19106326', 'cg19106353', '\n",
      "1000 input query terms found no hit:\t['cg19154695', 'cg19154698', 'cg19154834', 'cg19154858', 'cg19154861', 'cg19154940', 'cg19155007', '\n",
      "1000 input query terms found no hit:\t['cg19205831', 'cg19205847', 'cg19205850', 'cg19205897', 'cg19205909', 'cg19205926', 'cg19205994', '\n",
      "1000 input query terms found no hit:\t['cg19251242', 'cg19251280', 'cg19251291', 'cg19251310', 'cg19251326', 'cg19251345', 'cg19251352', '\n",
      "1000 input query terms found no hit:\t['cg19296429', 'cg19296486', 'cg19296501', 'cg19296504', 'cg19296509', 'cg19296571', 'cg19296591', '\n",
      "1000 input query terms found no hit:\t['cg19343783', 'cg19343809', 'cg19343871', 'cg19343880', 'cg19343912', 'cg19344013', 'cg19344082', '\n",
      "1000 input query terms found no hit:\t['cg19389641', 'cg19389763', 'cg19389852', 'cg19389875', 'cg19389884', 'cg19389944', 'cg19389953', '\n",
      "1000 input query terms found no hit:\t['cg19433211', 'cg19433225', 'cg19433305', 'cg19433341', 'cg19433435', 'cg19433487', 'cg19433697', '\n",
      "1000 input query terms found no hit:\t['cg19478558', 'cg19478606', 'cg19478638', 'cg19478678', 'cg19478682', 'cg19478709', 'cg19478711', '\n",
      "1000 input query terms found no hit:\t['cg19522185', 'cg19522202', 'cg19522294', 'cg19522332', 'cg19522467', 'cg19522565', 'cg19522644', '\n",
      "1000 input query terms found no hit:\t['cg19566395', 'cg19566405', 'cg19566417', 'cg19566482', 'cg19566641', 'cg19566648', 'cg19566663', '\n",
      "1000 input query terms found no hit:\t['cg19610971', 'cg19611002', 'cg19611013', 'cg19611068', 'cg19611103', 'cg19611163', 'cg19611175', '\n",
      "1000 input query terms found no hit:\t['cg19654118', 'cg19654142', 'cg19654195', 'cg19654209', 'cg19654224', 'cg19654437', 'cg19654586', '\n",
      "1000 input query terms found no hit:\t['cg19699832', 'cg19699893', 'cg19699911', 'cg19699962', 'cg19699973', 'cg19699994', 'cg19700054', '\n",
      "1000 input query terms found no hit:\t['cg19742470', 'cg19742538', 'cg19742623', 'cg19742686', 'cg19742687', 'cg19742717', 'cg19742789', '\n",
      "1000 input query terms found no hit:\t['cg19782173', 'cg19782190', 'cg19782202', 'cg19782242', 'cg19782243', 'cg19782322', 'cg19782411', '\n",
      "1000 input query terms found no hit:\t['cg19825240', 'cg19825277', 'cg19825302', 'cg19825305', 'cg19825353', 'cg19825371', 'cg19825437', '\n",
      "1000 input query terms found no hit:\t['cg19867410', 'cg19867445', 'cg19867476', 'cg19867481', 'cg19867579', 'cg19867615', 'cg19867649', '\n",
      "1000 input query terms found no hit:\t['cg19912142', 'cg19912160', 'cg19912197', 'cg19912283', 'cg19912309', 'cg19912332', 'cg19912337', '\n",
      "1000 input query terms found no hit:\t['cg19954334', 'cg19954341', 'cg19954348', 'cg19954357', 'cg19954374', 'cg19954448', 'cg19954481', '\n",
      "1000 input query terms found no hit:\t['cg20000049', 'cg20000099', 'cg20000107', 'cg20000129', 'cg20000132', 'cg20000185', 'cg20000220', '\n",
      "1000 input query terms found no hit:\t['cg20045841', 'cg20045888', 'cg20045910', 'cg20045970', 'cg20046003', 'cg20046101', 'cg20046126', '\n",
      "1000 input query terms found no hit:\t['cg20092348', 'cg20092356', 'cg20092376', 'cg20092396', 'cg20092399', 'cg20092458', 'cg20092482', '\n",
      "1000 input query terms found no hit:\t['cg20138404', 'cg20138429', 'cg20138458', 'cg20138489', 'cg20138604', 'cg20138617', 'cg20138695', '\n",
      "1000 input query terms found no hit:\t['cg20184893', 'cg20184917', 'cg20185005', 'cg20185017', 'cg20185083', 'cg20185106', 'cg20185145', '\n",
      "1000 input query terms found no hit:\t['cg20234121', 'cg20234151', 'cg20234170', 'cg20234205', 'cg20234266', 'cg20234280', 'cg20234282', '\n",
      "1000 input query terms found no hit:\t['cg20277616', 'cg20277670', 'cg20277781', 'cg20277872', 'cg20277905', 'cg20277909', 'cg20277977', '\n",
      "1000 input query terms found no hit:\t['cg20322853', 'cg20322862', 'cg20322876', 'cg20322901', 'cg20322977', 'cg20323015', 'cg20323072', '\n",
      "1000 input query terms found no hit:\t['cg20370152', 'cg20370281', 'cg20370293', 'cg20370296', 'cg20370364', 'cg20370505', 'cg20370568', '\n",
      "1000 input query terms found no hit:\t['cg20415459', 'cg20415486', 'cg20415517', 'cg20415529', 'cg20415600', 'cg20415643', 'cg20415809', '\n",
      "1000 input query terms found no hit:\t['cg20459256', 'cg20459327', 'cg20459332', 'cg20459379', 'cg20459407', 'cg20459430', 'cg20459495', '\n",
      "1000 input query terms found no hit:\t['cg20507224', 'cg20507228', 'cg20507256', 'cg20507276', 'cg20507363', 'cg20507401', 'cg20507414', '\n",
      "1000 input query terms found no hit:\t['cg20554943', 'cg20554948', 'cg20555018', 'cg20555040', 'cg20555138', 'cg20555222', 'cg20555234', '\n",
      "1000 input query terms found no hit:\t['cg20600034', 'cg20600051', 'cg20600053', 'cg20600099', 'cg20600130', 'cg20600205', 'cg20600210', '\n",
      "1000 input query terms found no hit:\t['cg20648333', 'cg20648350', 'cg20648385', 'cg20648497', 'cg20648501', 'cg20648530', 'cg20648539', '\n",
      "1000 input query terms found no hit:\t['cg20691722', 'cg20691781', 'cg20691837', 'cg20692011', 'cg20692029', 'cg20692059', 'cg20692149', '\n",
      "1000 input query terms found no hit:\t['cg20732690', 'cg20732755', 'cg20732763', 'cg20732787', 'cg20732876', 'cg20732881', 'cg20732955', '\n",
      "1000 input query terms found no hit:\t['cg20779964', 'cg20779966', 'cg20780013', 'cg20780105', 'cg20780131', 'cg20780180', 'cg20780302', '\n",
      "1000 input query terms found no hit:\t['cg20825216', 'cg20825221', 'cg20825307', 'cg20825355', 'cg20825416', 'cg20825428', 'cg20825437', '\n",
      "1000 input query terms found no hit:\t['cg20876061', 'cg20876117', 'cg20876166', 'cg20876170', 'cg20876215', 'cg20876248', 'cg20876266', '\n",
      "1000 input query terms found no hit:\t['cg20923753', 'cg20923754', 'cg20923800', 'cg20923877', 'cg20923885', 'cg20924082', 'cg20924136', '\n",
      "1000 input query terms found no hit:\t['cg20967802', 'cg20967818', 'cg20967819', 'cg20967869', 'cg20967887', 'cg20967889', 'cg20967975', '\n",
      "1000 input query terms found no hit:\t['cg21010925', 'cg21010995', 'cg21011025', 'cg21011036', 'cg21011043', 'cg21011066', 'cg21011110', '\n",
      "1000 input query terms found no hit:\t['cg21054473', 'cg21054521', 'cg21054550', 'cg21054601', 'cg21054655', 'cg21054693', 'cg21054703', '\n",
      "1000 input query terms found no hit:\t['cg21103383', 'cg21103385', 'cg21103400', 'cg21103567', 'cg21103587', 'cg21103728', 'cg21103734', '\n",
      "1000 input query terms found no hit:\t['cg21149843', 'cg21149863', 'cg21149944', 'cg21149967', 'cg21149983', 'cg21150003', 'cg21150026', '\n",
      "1000 input query terms found no hit:\t['cg21188989', 'cg21189054', 'cg21189063', 'cg21189099', 'cg21189101', 'cg21189146', 'cg21189154', '\n",
      "1000 input query terms found no hit:\t['cg21229278', 'cg21229297', 'cg21229337', 'cg21229339', 'cg21229340', 'cg21229396', 'cg21229482', '\n",
      "1000 input query terms found no hit:\t['cg21272664', 'cg21272713', 'cg21272716', 'cg21272795', 'cg21272843', 'cg21272866', 'cg21272882', '\n",
      "1000 input query terms found no hit:\t['cg21320455', 'cg21320567', 'cg21320569', 'cg21320573', 'cg21320638', 'cg21320768', 'cg21320793', '\n",
      "1000 input query terms found no hit:\t['cg21372451', 'cg21372533', 'cg21372571', 'cg21372733', 'cg21372755', 'cg21372810', 'cg21372885', '\n",
      "1000 input query terms found no hit:\t['cg21421331', 'cg21421408', 'cg21421501', 'cg21421513', 'cg21421618', 'cg21421634', 'cg21421692', '\n",
      "1000 input query terms found no hit:\t['cg21467616', 'cg21467631', 'cg21467684', 'cg21467692', 'cg21467697', 'cg21467717', 'cg21467719', '\n",
      "1000 input query terms found no hit:\t['cg21511972', 'cg21511976', 'cg21511979', 'cg21511981', 'cg21512051', 'cg21512075', 'cg21512124', '\n",
      "1000 input query terms found no hit:\t['cg21554845', 'cg21554861', 'cg21554886', 'cg21554895', 'cg21554922', 'cg21554959', 'cg21555085', '\n",
      "1000 input query terms found no hit:\t['cg21597685', 'cg21597692', 'cg21597712', 'cg21597754', 'cg21597760', 'cg21597937', 'cg21598040', '\n",
      "1000 input query terms found no hit:\t['cg21643122', 'cg21643128', 'cg21643150', 'cg21643178', 'cg21643191', 'cg21643217', 'cg21643262', '\n",
      "1000 input query terms found no hit:\t['cg21690982', 'cg21691014', 'cg21691063', 'cg21691065', 'cg21691089', 'cg21691116', 'cg21691161', '\n",
      "1000 input query terms found no hit:\t['cg21740826', 'cg21740925', 'cg21740943', 'cg21740960', 'cg21740964', 'cg21741030', 'cg21741081', '\n",
      "1000 input query terms found no hit:\t['cg21786295', 'cg21786334', 'cg21786349', 'cg21786381', 'cg21786386', 'cg21786399', 'cg21786459', '\n",
      "1000 input query terms found no hit:\t['cg21832073', 'cg21832091', 'cg21832096', 'cg21832142', 'cg21832144', 'cg21832150', 'cg21832294', '\n",
      "1000 input query terms found no hit:\t['cg21875815', 'cg21875839', 'cg21875845', 'cg21875866', 'cg21875946', 'cg21875948', 'cg21875980', '\n",
      "1000 input query terms found no hit:\t['cg21919596', 'cg21919599', 'cg21919602', 'cg21919718', 'cg21919719', 'cg21919729', 'cg21919766', '\n",
      "1000 input query terms found no hit:\t['cg21965723', 'cg21965820', 'cg21965843', 'cg21965980', 'cg21965995', 'cg21965997', 'cg21966075', '\n",
      "1000 input query terms found no hit:\t['cg22016094', 'cg22016170', 'cg22016187', 'cg22016202', 'cg22016288', 'cg22016326', 'cg22016357', '\n",
      "1000 input query terms found no hit:\t['cg22060113', 'cg22060130', 'cg22060153', 'cg22060161', 'cg22060227', 'cg22060423', 'cg22060484', '\n",
      "1000 input query terms found no hit:\t['cg22108941', 'cg22108980', 'cg22109056', 'cg22109165', 'cg22109234', 'cg22109240', 'cg22109262', '\n",
      "1000 input query terms found no hit:\t['cg22153527', 'cg22153695', 'cg22153696', 'cg22153708', 'cg22153745', 'cg22153799', 'cg22153823', '\n",
      "1000 input query terms found no hit:\t['cg22202181', 'cg22202182', 'cg22202191', 'cg22202234', 'cg22202345', 'cg22202381', 'cg22202401', '\n",
      "1000 input query terms found no hit:\t['cg22249932', 'cg22249956', 'cg22249985', 'cg22250145', 'cg22250168', 'cg22250192', 'cg22250242', '\n",
      "1000 input query terms found no hit:\t['cg22298183', 'cg22298206', 'cg22298208', 'cg22298224', 'cg22298236', 'cg22298265', 'cg22298418', '\n",
      "1000 input query terms found no hit:\t['cg22346261', 'cg22346297', 'cg22346355', 'cg22346356', 'cg22346380', 'cg22346429', 'cg22346461', '\n",
      "1000 input query terms found no hit:\t['cg22391888', 'cg22391906', 'cg22391913', 'cg22392029', 'cg22392046', 'cg22392178', 'cg22392250', '\n",
      "1000 input query terms found no hit:\t['cg22441615', 'cg22441683', 'cg22441758', 'cg22441770', 'cg22441810', 'cg22441846', 'cg22441882', '\n",
      "1000 input query terms found no hit:\t['cg22489878', 'cg22489883', 'cg22489931', 'cg22489957', 'cg22490000', 'cg22490084', 'cg22490134', '\n",
      "1000 input query terms found no hit:\t['cg22534206', 'cg22534231', 'cg22534288', 'cg22534310', 'cg22534336', 'cg22534339', 'cg22534374', '\n",
      "1000 input query terms found no hit:\t['cg22580670', 'cg22580680', 'cg22580712', 'cg22580846', 'cg22580900', 'cg22580903', 'cg22580932', '\n",
      "1000 input query terms found no hit:\t['cg22627753', 'cg22627800', 'cg22627826', 'cg22627866', 'cg22627876', 'cg22627940', 'cg22627950', '\n",
      "1000 input query terms found no hit:\t['cg22675911', 'cg22675937', 'cg22675956', 'cg22675970', 'cg22675981', 'cg22676000', 'cg22676023', '\n",
      "1000 input query terms found no hit:\t['cg22717389', 'cg22717461', 'cg22717476', 'cg22717593', 'cg22717607', 'cg22717608', 'cg22717723', '\n",
      "1000 input query terms found no hit:\t['cg22764186', 'cg22764193', 'cg22764231', 'cg22764285', 'cg22764289', 'cg22764338', 'cg22764341', '\n",
      "1000 input query terms found no hit:\t['cg22807378', 'cg22807402', 'cg22807436', 'cg22807449', 'cg22807516', 'cg22807519', 'cg22807551', '\n",
      "1000 input query terms found no hit:\t['cg22851990', 'cg22852013', 'cg22852065', 'cg22852103', 'cg22852149', 'cg22852177', 'cg22852191', '\n",
      "1000 input query terms found no hit:\t['cg22891988', 'cg22892036', 'cg22892043', 'cg22892110', 'cg22892154', 'cg22892201', 'cg22892237', '\n",
      "1000 input query terms found no hit:\t['cg22940525', 'cg22940575', 'cg22940675', 'cg22940738', 'cg22940798', 'cg22940815', 'cg22940816', '\n",
      "1000 input query terms found no hit:\t['cg22986737', 'cg22986770', 'cg22986786', 'cg22986793', 'cg22986872', 'cg22986925', 'cg22986941', '\n",
      "1000 input query terms found no hit:\t['cg23029041', 'cg23029100', 'cg23029159', 'cg23029183', 'cg23029193', 'cg23029198', 'cg23029228', '\n",
      "1000 input query terms found no hit:\t['cg23070026', 'cg23070073', 'cg23070111', 'cg23070153', 'cg23070249', 'cg23070297', 'cg23070487', '\n",
      "1000 input query terms found no hit:\t['cg23115387', 'cg23115390', 'cg23115560', 'cg23115674', 'cg23115907', 'cg23115933', 'cg23115987', '\n",
      "1000 input query terms found no hit:\t['cg23163999', 'cg23164054', 'cg23164076', 'cg23164083', 'cg23164183', 'cg23164203', 'cg23164218', '\n",
      "1000 input query terms found no hit:\t['cg23207527', 'cg23207542', 'cg23207597', 'cg23207616', 'cg23207621', 'cg23207673', 'cg23207676', '\n",
      "1000 input query terms found no hit:\t['cg23249140', 'cg23249207', 'cg23249239', 'cg23249254', 'cg23249278', 'cg23249321', 'cg23249336', '\n",
      "1000 input query terms found no hit:\t['cg23295647', 'cg23295754', 'cg23295826', 'cg23295886', 'cg23295942', 'cg23295955', 'cg23295991', '\n",
      "1000 input query terms found no hit:\t['cg23341202', 'cg23341223', 'cg23341351', 'cg23341459', 'cg23341486', 'cg23341488', 'cg23341504', '\n",
      "1000 input query terms found no hit:\t['cg23384428', 'cg23384538', 'cg23384615', 'cg23384668', 'cg23384688', 'cg23384708', 'cg23384737', '\n",
      "1000 input query terms found no hit:\t['cg23430644', 'cg23430652', 'cg23430666', 'cg23430742', 'cg23430771', 'cg23430783', 'cg23431342', '\n",
      "1000 input query terms found no hit:\t['cg23475710', 'cg23475725', 'cg23475732', 'cg23475755', 'cg23475778', 'cg23475787', 'cg23475858', '\n",
      "1000 input query terms found no hit:\t['cg23519572', 'cg23519626', 'cg23519631', 'cg23519637', 'cg23519671', 'cg23519673', 'cg23519755', '\n",
      "1000 input query terms found no hit:\t['cg23563927', 'cg23563958', 'cg23563960', 'cg23564005', 'cg23564028', 'cg23564178', 'cg23564228', '\n",
      "1000 input query terms found no hit:\t['cg23610575', 'cg23610588', 'cg23610595', 'cg23610682', 'cg23610736', 'cg23610841', 'cg23610994', '\n",
      "1000 input query terms found no hit:\t['cg23652681', 'cg23652725', 'cg23652734', 'cg23652770', 'cg23652785', 'cg23652814', 'cg23652840', '\n",
      "1000 input query terms found no hit:\t['cg23689865', 'cg23689916', 'cg23689925', 'cg23689985', 'cg23690033', 'cg23690294', 'cg23690302', '\n",
      "1000 input query terms found no hit:\t['cg23732024', 'cg23732049', 'cg23732080', 'cg23732138', 'cg23732147', 'cg23732169', 'cg23732346', '\n",
      "1000 input query terms found no hit:\t['cg23782490', 'cg23782535', 'cg23782616', 'cg23782650', 'cg23782689', 'cg23782719', 'cg23782734', '\n",
      "1000 input query terms found no hit:\t['cg23831217', 'cg23831328', 'cg23831366', 'cg23831386', 'cg23831424', 'cg23831458', 'cg23831536', '\n",
      "1000 input query terms found no hit:\t['cg23875532', 'cg23875597', 'cg23875752', 'cg23875758', 'cg23875821', 'cg23875854', 'cg23875891', '\n",
      "1000 input query terms found no hit:\t['cg23916878', 'cg23916896', 'cg23916948', 'cg23916969', 'cg23916976', 'cg23916979', 'cg23917009', '\n",
      "1000 input query terms found no hit:\t['cg23959119', 'cg23959187', 'cg23959206', 'cg23959311', 'cg23959469', 'cg23959474', 'cg23959518', '\n",
      "1000 input query terms found no hit:\t['cg24000241', 'cg24000259', 'cg24000285', 'cg24000300', 'cg24000316', 'cg24000333', 'cg24000344', '\n",
      "1000 input query terms found no hit:\t['cg24042313', 'cg24042402', 'cg24042452', 'cg24042502', 'cg24042517', 'cg24042528', 'cg24042578', '\n",
      "1000 input query terms found no hit:\t['cg24086717', 'cg24086751', 'cg24086758', 'cg24086845', 'cg24086869', 'cg24086889', 'cg24086920', '\n",
      "1000 input query terms found no hit:\t['cg24130348', 'cg24130365', 'cg24130392', 'cg24130424', 'cg24130459', 'cg24130476', 'cg24130494', '\n",
      "1000 input query terms found no hit:\t['cg24172269', 'cg24172278', 'cg24172291', 'cg24172324', 'cg24172509', 'cg24172523', 'cg24172530', '\n",
      "1000 input query terms found no hit:\t['cg24219671', 'cg24219697', 'cg24219715', 'cg24219732', 'cg24219734', 'cg24219777', 'cg24219822', '\n",
      "1000 input query terms found no hit:\t['cg24270398', 'cg24270471', 'cg24270542', 'cg24270596', 'cg24270629', 'cg24270678', 'cg24270700', '\n",
      "1000 input query terms found no hit:\t['cg24315049', 'cg24315188', 'cg24315197', 'cg24315209', 'cg24315305', 'cg24315340', 'cg24315421', '\n",
      "1000 input query terms found no hit:\t['cg24360034', 'cg24360069', 'cg24360099', 'cg24360131', 'cg24360137', 'cg24360174', 'cg24360197', '\n",
      "1000 input query terms found no hit:\t['cg24403338', 'cg24403390', 'cg24403394', 'cg24403479', 'cg24403487', 'cg24403497', 'cg24403549', '\n",
      "1000 input query terms found no hit:\t['cg24446457', 'cg24446548', 'cg24446586', 'cg24446604', 'cg24446610', 'cg24446660', 'cg24446823', '\n",
      "1000 input query terms found no hit:\t['cg24491028', 'cg24491078', 'cg24491328', 'cg24491364', 'cg24491369', 'cg24491493', 'cg24491553', '\n",
      "1000 input query terms found no hit:\t['cg24531825', 'cg24531915', 'cg24531949', 'cg24531955', 'cg24531977', 'cg24531993', 'cg24532011', '\n",
      "1000 input query terms found no hit:\t['cg24579212', 'cg24579218', 'cg24579224', 'cg24579246', 'cg24579247', 'cg24579258', 'cg24579320', '\n",
      "1000 input query terms found no hit:\t['cg24623326', 'cg24623360', 'cg24623368', 'cg24623376', 'cg24623382', 'cg24623422', 'cg24623434', '\n",
      "1000 input query terms found no hit:\t['cg24668728', 'cg24668736', 'cg24668753', 'cg24668836', 'cg24668854', 'cg24668859', 'cg24668869', '\n",
      "1000 input query terms found no hit:\t['cg24708350', 'cg24708354', 'cg24708471', 'cg24708480', 'cg24708518', 'cg24708545', 'cg24708568', '\n",
      "1000 input query terms found no hit:\t['cg24749455', 'cg24749470', 'cg24749475', 'cg24749559', 'cg24749602', 'cg24749636', 'cg24749656', '\n",
      "1000 input query terms found no hit:\t['cg24794636', 'cg24794669', 'cg24794697', 'cg24794709', 'cg24794734', 'cg24794758', 'cg24794760', '\n",
      "1000 input query terms found no hit:\t['cg24838839', 'cg24838864', 'cg24838897', 'cg24838949', 'cg24838977', 'cg24839039', 'cg24839145', '\n",
      "1000 input query terms found no hit:\t['cg24878607', 'cg24878631', 'cg24878755', 'cg24878791', 'cg24878868', 'cg24878887', 'cg24878986', '\n",
      "1000 input query terms found no hit:\t['cg24925313', 'cg24925345', 'cg24925400', 'cg24925546', 'cg24925593', 'cg24925635', 'cg24925696', '\n",
      "1000 input query terms found no hit:\t['cg24970361', 'cg24970403', 'cg24970449', 'cg24970558', 'cg24970620', 'cg24970879', 'cg24970918', '\n",
      "1000 input query terms found no hit:\t['cg25014118', 'cg25014180', 'cg25014201', 'cg25014222', 'cg25014224', 'cg25014228', 'cg25014229', '\n",
      "1000 input query terms found no hit:\t['cg25055403', 'cg25055433', 'cg25055477', 'cg25055520', 'cg25055576', 'cg25055656', 'cg25055744', '\n",
      "1000 input query terms found no hit:\t['cg25103379', 'cg25103393', 'cg25103417', 'cg25103419', 'cg25103492', 'cg25103559', 'cg25103650', '\n",
      "1000 input query terms found no hit:\t['cg25146508', 'cg25146526', 'cg25146557', 'cg25146575', 'cg25146577', 'cg25146597', 'cg25146728', '\n",
      "1000 input query terms found no hit:\t['cg25194502', 'cg25194612', 'cg25194694', 'cg25194720', 'cg25194807', 'cg25194841', 'cg25194918', '\n",
      "1000 input query terms found no hit:\t['cg25237552', 'cg25237566', 'cg25237646', 'cg25237692', 'cg25237720', 'cg25237755', 'cg25237788', '\n",
      "1000 input query terms found no hit:\t['cg25285631', 'cg25285646', 'cg25285658', 'cg25285743', 'cg25285762', 'cg25285794', 'cg25285805', '\n",
      "1000 input query terms found no hit:\t['cg25330514', 'cg25330665', 'cg25330725', 'cg25330773', 'cg25330843', 'cg25330878', 'cg25330919', '\n",
      "1000 input query terms found no hit:\t['cg25373595', 'cg25373620', 'cg25373624', 'cg25373630', 'cg25373646', 'cg25373696', 'cg25373746', '\n",
      "1000 input query terms found no hit:\t['cg25418309', 'cg25418363', 'cg25418380', 'cg25418381', 'cg25418406', 'cg25418474', 'cg25418481', '\n",
      "1000 input query terms found no hit:\t['cg25462573', 'cg25462669', 'cg25462716', 'cg25462718', 'cg25462790', 'cg25462795', 'cg25462815', '\n",
      "1000 input query terms found no hit:\t['cg25505570', 'cg25505605', 'cg25505610', 'cg25505611', 'cg25505676', 'cg25505729', 'cg25505785', '\n",
      "1000 input query terms found no hit:\t['cg25554117', 'cg25554152', 'cg25554183', 'cg25554190', 'cg25554193', 'cg25554205', 'cg25554237', '\n",
      "1000 input query terms found no hit:\t['cg25596675', 'cg25596709', 'cg25596745', 'cg25596754', 'cg25596756', 'cg25596783', 'cg25596796', '\n",
      "1000 input query terms found no hit:\t['cg25639595', 'cg25639601', 'cg25639671', 'cg25639749', 'cg25639809', 'cg25639848', 'cg25639858', '\n",
      "1000 input query terms found no hit:\t['cg25687507', 'cg25687573', 'cg25687585', 'cg25687744', 'cg25687777', 'cg25687781', 'cg25687782', '\n",
      "1000 input query terms found no hit:\t['cg25732426', 'cg25732446', 'cg25732462', 'cg25732522', 'cg25732612', 'cg25732668', 'cg25732687', '\n",
      "1000 input query terms found no hit:\t['cg25773917', 'cg25773922', 'cg25773935', 'cg25774020', 'cg25774094', 'cg25774116', 'cg25774148', '\n",
      "1000 input query terms found no hit:\t['cg25820858', 'cg25820914', 'cg25820971', 'cg25820993', 'cg25821040', 'cg25821072', 'cg25821138', '\n",
      "1000 input query terms found no hit:\t['cg25869250', 'cg25869267', 'cg25869295', 'cg25869317', 'cg25869578', 'cg25869635', 'cg25869732', '\n",
      "1000 input query terms found no hit:\t['cg25920814', 'cg25920850', 'cg25920857', 'cg25920938', 'cg25921007', 'cg25921044', 'cg25921170', '\n",
      "1000 input query terms found no hit:\t['cg25964035', 'cg25964085', 'cg25964141', 'cg25964180', 'cg25964213', 'cg25964253', 'cg25964299', '\n",
      "1000 input query terms found no hit:\t['cg26009013', 'cg26009029', 'cg26009035', 'cg26009102', 'cg26009124', 'cg26009180', 'cg26009192', '\n",
      "1000 input query terms found no hit:\t['cg26055812', 'cg26055840', 'cg26055841', 'cg26055894', 'cg26055899', 'cg26055923', 'cg26055948', '\n",
      "1000 input query terms found no hit:\t['cg26099468', 'cg26099549', 'cg26099570', 'cg26099607', 'cg26099633', 'cg26099766', 'cg26099778', '\n",
      "1000 input query terms found no hit:\t['cg26141324', 'cg26141335', 'cg26141336', 'cg26141385', 'cg26141494', 'cg26141510', 'cg26141526', '\n",
      "1000 input query terms found no hit:\t['cg26185180', 'cg26185198', 'cg26185225', 'cg26185340', 'cg26185347', 'cg26185356', 'cg26185392', '\n",
      "1000 input query terms found no hit:\t['cg26227523', 'cg26227569', 'cg26227592', 'cg26227627', 'cg26227661', 'cg26227773', 'cg26227840', '\n",
      "1000 input query terms found no hit:\t['cg26270725', 'cg26270738', 'cg26270746', 'cg26270776', 'cg26270792', 'cg26270867', 'cg26270879', '\n",
      "1000 input query terms found no hit:\t['cg26311734', 'cg26311781', 'cg26311782', 'cg26311804', 'cg26311932', 'cg26311944', 'cg26311995', '\n",
      "1000 input query terms found no hit:\t['cg26350759', 'cg26350784', 'cg26350803', 'cg26350875', 'cg26350878', 'cg26350916', 'cg26350919', '\n",
      "1000 input query terms found no hit:\t['cg26395066', 'cg26395071', 'cg26395089', 'cg26395114', 'cg26395115', 'cg26395193', 'cg26395211', '\n",
      "1000 input query terms found no hit:\t['cg26438105', 'cg26438167', 'cg26438181', 'cg26438215', 'cg26438234', 'cg26438284', 'cg26438294', '\n",
      "1000 input query terms found no hit:\t['cg26480069', 'cg26480091', 'cg26480092', 'cg26480227', 'cg26480228', 'cg26480358', 'cg26480387', '\n",
      "1000 input query terms found no hit:\t['cg26524899', 'cg26524945', 'cg26524964', 'cg26525008', 'cg26525025', 'cg26525074', 'cg26525076', '\n",
      "1000 input query terms found no hit:\t['cg26565544', 'cg26565578', 'cg26565580', 'cg26565603', 'cg26565626', 'cg26565649', 'cg26565676', '\n",
      "1000 input query terms found no hit:\t['cg26611723', 'cg26611765', 'cg26611809', 'cg26611815', 'cg26611832', 'cg26611836', 'cg26611911', '\n",
      "1000 input query terms found no hit:\t['cg26655028', 'cg26655044', 'cg26655076', 'cg26655085', 'cg26655130', 'cg26655143', 'cg26655226', '\n",
      "1000 input query terms found no hit:\t['cg26693616', 'cg26693641', 'cg26693679', 'cg26693693', 'cg26693701', 'cg26693760', 'cg26693817', '\n",
      "1000 input query terms found no hit:\t['cg26737305', 'cg26737313', 'cg26737357', 'cg26737458', 'cg26737496', 'cg26737520', 'cg26737542', '\n",
      "1000 input query terms found no hit:\t['cg26783622', 'cg26783666', 'cg26783679', 'cg26783691', 'cg26783727', 'cg26784000', 'cg26784011', '\n",
      "1000 input query terms found no hit:\t['cg26826077', 'cg26826109', 'cg26826117', 'cg26826136', 'cg26826142', 'cg26826183', 'cg26826207', '\n",
      "1000 input query terms found no hit:\t['cg26867978', 'cg26867987', 'cg26868023', 'cg26868097', 'cg26868111', 'cg26868156', 'cg26868265', '\n",
      "1000 input query terms found no hit:\t['cg26909184', 'cg26909217', 'cg26909237', 'cg26909249', 'cg26909276', 'cg26909299', 'cg26909431', '\n",
      "1000 input query terms found no hit:\t['cg26950332', 'cg26950369', 'cg26950463', 'cg26950475', 'cg26950531', 'cg26950557', 'cg26950560', '\n",
      "1000 input query terms found no hit:\t['cg26992244', 'cg26992245', 'cg26992381', 'cg26992387', 'cg26992415', 'cg26992486', 'cg26992539', '\n",
      "1000 input query terms found no hit:\t['cg27036347', 'cg27036376', 'cg27036456', 'cg27036462', 'cg27036581', 'cg27036638', 'cg27036733', '\n",
      "1000 input query terms found no hit:\t['cg27078562', 'cg27078600', 'cg27078632', 'cg27078652', 'cg27078729', 'cg27078786', 'cg27078812', '\n",
      "1000 input query terms found no hit:\t['cg27119296', 'cg27119318', 'cg27119358', 'cg27119392', 'cg27119412', 'cg27119415', 'cg27119428', '\n",
      "1000 input query terms found no hit:\t['cg27161973', 'cg27162030', 'cg27162122', 'cg27162286', 'cg27162375', 'cg27162435', 'cg27162464', '\n",
      "1000 input query terms found no hit:\t['cg27208169', 'cg27208248', 'cg27208256', 'cg27208307', 'cg27208362', 'cg27208536', 'cg27208573', '\n",
      "1000 input query terms found no hit:\t['cg27253887', 'cg27254046', 'cg27254056', 'cg27254068', 'cg27254118', 'cg27254295', 'cg27254338', '\n",
      "1000 input query terms found no hit:\t['cg27297221', 'cg27297263', 'cg27297315', 'cg27297318', 'cg27297376', 'cg27297434', 'cg27297505', '\n",
      "1000 input query terms found no hit:\t['cg27334899', 'cg27334900', 'cg27334906', 'cg27334913', 'cg27334919', 'cg27334928', 'cg27334938', '\n",
      "1000 input query terms found no hit:\t['cg27377863', 'cg27377960', 'cg27378015', 'cg27378029', 'cg27378035', 'cg27378180', 'cg27378200', '\n",
      "1000 input query terms found no hit:\t['cg27420273', 'cg27420361', 'cg27420365', 'cg27420415', 'cg27420520', 'cg27420558', 'cg27420590', '\n",
      "1000 input query terms found no hit:\t['cg27462969', 'cg27462975', 'cg27462988', 'cg27463004', 'cg27463141', 'cg27463181', 'cg27463246', '\n",
      "1000 input query terms found no hit:\t['cg27506503', 'cg27506515', 'cg27506533', 'cg27506547', 'cg27506609', 'cg27506783', 'cg27506810', '\n",
      "1000 input query terms found no hit:\t['cg27547132', 'cg27547165', 'cg27547269', 'cg27547272', 'cg27547281', 'cg27547282', 'cg27547283', '\n",
      "1000 input query terms found no hit:\t['cg27589058', 'cg27589088', 'cg27589105', 'cg27589196', 'cg27589204', 'cg27589230', 'cg27589276', '\n",
      "822 input query terms found no hit:\t['cg27632677', 'cg27632696', 'cg27632704', 'cg27632817', 'cg27632861', 'cg27632891', 'cg27632901', '\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 62 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Poor Metabolizer of Drugs', the least common label is '0.0' with 10 occurrences. This represents 16.13% of the dataset.\n",
      "The distribution of the feature 'Poor Metabolizer of Drugs' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "    \n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import json\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE155985</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>62</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "0  GSE155985       True          True      False    False       False   \n",
       "\n",
       "   sample_size note  \n",
       "0           62       "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.295070023Z",
     "start_time": "2024-01-10T21:40:21.338515691Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +: 'NoneType' and 'str'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [53]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m merged_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, \u001b[43mcondition_best_cohort\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m))\n\u001b[1;32m      2\u001b[0m merged_data\u001b[38;5;241m.\u001b[39mhead()\n",
      "\u001b[0;31mTypeError\u001b[0m: unsupported operand type(s) for +: 'NoneType' and 'str'"
     ]
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.358779256Z",
     "start_time": "2024-01-10T21:40:27.271337819Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')\n",
    "\n",
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_batch_effect(X):\n",
    "    \"\"\"\n",
    "    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n",
    "\n",
    "    Args:\n",
    "    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n",
    "\n",
    "    Returns:\n",
    "    bool: True if a potential batch effect is detected, False otherwise.\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "\n",
    "    # Computing XX^T\n",
    "    XXt = np.dot(X, X.T)\n",
    "\n",
    "    # Compute the eigenvalues of XX^T\n",
    "    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n",
    "    eigen_values = sorted(eigen_values, reverse=True)\n",
    "\n",
    "    # Check for large gaps in the eigenvalues\n",
    "    for i in range(len(eigen_values) - 1):\n",
    "        gap = eigen_values[i] - eigen_values[i + 1]\n",
    "        if gap > 1 / n_samples:  # You may need to adjust this threshold\n",
    "            return True\n",
    "\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546590575Z",
     "start_time": "2024-01-10T21:40:27.298495045Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sparse_lmm import VariableSelection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:27.546733889Z",
     "start_time": "2024-01-10T21:40:27.535249532Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Select appropriate models based on whether the dataset has batch effect.\n",
    "# We experiment on two models for each branch. We will decide which one to choose later.\n",
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n",
    "    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n",
    "    indices = np.arange(X.shape[0])\n",
    "    np.random.shuffle(indices)\n",
    "\n",
    "    fold_size = len(X) // k\n",
    "    performances = []\n",
    "\n",
    "    for i in range(k):\n",
    "        # Split data into train and test based on the current fold\n",
    "        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n",
    "        train_indices = np.setdiff1d(indices, test_indices)\n",
    "\n",
    "        X_train, X_test = X[train_indices], X[test_indices]\n",
    "        Y_train, Y_test = Y[train_indices], Y[test_indices]\n",
    "        Z_train, Z_test = Z[train_indices], Z[test_indices]\n",
    "\n",
    "        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n",
    "        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n",
    "\n",
    "        # model = model_constructor(**model_params)\n",
    "        model = ResidualizationRegressor(model_constructor, model_params)\n",
    "        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n",
    "        predictions = model.predict(normalized_X_test, normalized_Z_test)\n",
    "\n",
    "        if target_type == 'binary':\n",
    "            predictions = (predictions > 0.5).astype(int)\n",
    "            Y_test = (Y_test > 0.5).astype(int)\n",
    "            performance = accuracy_score(Y_test, predictions)\n",
    "        elif target_type == 'continuous':\n",
    "            performance = mean_squared_error(Y_test, predictions)\n",
    "\n",
    "        performances.append(performance)\n",
    "\n",
    "    cv_mean = np.mean(performances)\n",
    "    cv_std = np.std(performances)\n",
    "\n",
    "    if target_type == 'binary':\n",
    "        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n",
    "    else:\n",
    "        print(f'The cross-validation MSE is {(cv_mean * 100):.2f} ± {(cv_std * 100):.2f}')\n",
    "\n",
    "    return cv_mean, cv_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_data(X_train, X_test=None):\n",
    "    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n",
    "    and optionally the test data\"\"\"\n",
    "    mean = np.mean(X_train, axis=0)\n",
    "    std = np.std(X_train, axis=0)\n",
    "\n",
    "    # Handling columns with std = 0\n",
    "    std_no_zero = np.where(std == 0, 1, std)\n",
    "\n",
    "    # Normalize X_train\n",
    "    X_train_normalized = (X_train - mean) / std_no_zero\n",
    "    # Set normalized values to 0 where std was 0\n",
    "    X_train_normalized[:, std == 0] = 0\n",
    "\n",
    "    if X_test is not None:\n",
    "        X_test_normalized = (X_test - mean) / std_no_zero\n",
    "        X_test_normalized[:, std == 0] = 0\n",
    "    else:\n",
    "        X_test_normalized = None\n",
    "\n",
    "    return X_train_normalized, X_test_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ResidualizationRegressor:\n",
    "    def __init__(self, regression_model_constructor, params=None):\n",
    "        if params is None:\n",
    "            params = {}\n",
    "        self.regression_model = regression_model_constructor(**params)\n",
    "        self.beta_Z = None  # Coefficients for regression of Y on Z\n",
    "        self.beta_X = None  # Coefficients for regression of residual on X\n",
    "        self.neg_log_p_values = None  # Negative logarithm of p-values\n",
    "        self.p_values = None  # Actual p-values\n",
    "\n",
    "    def _reshape_data(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the data to ensure it's in the correct format (2D array).\n",
    "\n",
    "        :param data: The input data (can be 1D or 2D array).\n",
    "        :return: Reshaped 2D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 1:\n",
    "            return data.reshape(-1, 1)\n",
    "        return data\n",
    "\n",
    "    def _reshape_output(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the output data to ensure it's in the correct format (1D array).\n",
    "\n",
    "        :param data: The output data (can be 1D or 2D array).\n",
    "        :return: Reshaped 1D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 2 and data.shape[1] == 1:\n",
    "            return data.ravel()\n",
    "        return data\n",
    "    \n",
    "    def fit(self, X, Y, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Y = self._reshape_data(Y)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        # Step 1: Linear regression of Y on Z\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n",
    "        Y_hat = Z_ones @ self.beta_Z\n",
    "        e_Y = Y - Y_hat  # Residual of Y\n",
    "\n",
    "        # Step 2: Regress the residual on X using the included regression model\n",
    "        self.regression_model.fit(X, e_Y)\n",
    "\n",
    "        # Obtain coefficients from the regression model\n",
    "        if hasattr(self.regression_model, 'coef_'):\n",
    "            self.beta_X = self.regression_model.coef_\n",
    "        elif hasattr(self.regression_model, 'getBeta'):\n",
    "            beta_output = self.regression_model.getBeta()\n",
    "            self.beta_X = self._reshape_output(beta_output)\n",
    "\n",
    "        # Obtain negative logarithm of p-values, if available\n",
    "        if hasattr(self.regression_model, 'getNegLogP'):\n",
    "            neg_log_p_output = self.regression_model.getNegLogP()\n",
    "            if neg_log_p_output is not None:\n",
    "                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n",
    "                self.p_values = np.exp(-self.neg_log_p_values)\n",
    "                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n",
    "                p_values_Z = np.full(Z.shape[1], np.nan)\n",
    "                self.p_values = np.concatenate((p_values_Z, self.p_values))\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score, mean_squared_error\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:30.570471636Z",
     "start_time": "2024-01-10T21:40:27.546295159Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:47.575878022Z",
     "start_time": "2024-01-10T21:40:30.570907992Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.663737008Z",
     "start_time": "2024-01-10T21:40:47.576021186Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n",
    "                     threshold: float = 0.05, save_output: bool = True,\n",
    "                     output_dir: str = './output', model_id: int = 1) -> None:\n",
    "    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n",
    "    consists of one variable about condition and multiple variables about genetic factors.\n",
    "    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n",
    "    p-values or non-zero coefficients, depending on the availability of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    model (Any): The trained regression Model.\n",
    "    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n",
    "    trait (str): The target trait of interest.\n",
    "    condition (str): The specific condition to examine within the model.\n",
    "    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n",
    "    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n",
    "    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n",
    "    model_id (int): The index of the model, 1 or 2.\n",
    "\n",
    "    Returns:\n",
    "    None: This function does not return anything but prints and optionally saves the output.\n",
    "    \"\"\"\n",
    "    coefficients = model.get_coefficients().reshape(-1).tolist()\n",
    "    p_values = model.get_p_values()\n",
    "    if p_values is None:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients\n",
    "        })\n",
    "    else:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients,\n",
    "            'p_value': p_values.reshape(-1).tolist()\n",
    "        })\n",
    "    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n",
    "\n",
    "    print(f\"Effect of the condition on the target variable:\")\n",
    "    print(f\"Variable: {condition}\")\n",
    "    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n",
    "    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n",
    "    if p_values is None:\n",
    "        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n",
    "        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n",
    "        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n",
    "            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n",
    "    else:\n",
    "        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n",
    "        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n",
    "        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n",
    "        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n",
    "        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n",
    "            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n",
    "\n",
    "    print(significant_genes_sorted.to_string(index=False))\n",
    "\n",
    "    # Optionally, save this to a CSV file\n",
    "    if save_output:\n",
    "        significant_genes_sorted.to_csv(\n",
    "            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.673565176Z",
     "start_time": "2024-01-10T21:40:51.665291317Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:51.730825913Z",
     "start_time": "2024-01-10T21:40:51.675049728Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
