{"cells":[{"cell_type":"markdown","metadata":{"collapsed":false,"id":"NB8dJaS7ZWQI"},"source":["# Gold standard curation: Preprocessing and single-step regression"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"T8BiY3jEZWQL"},"source":["In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n","\n","1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n","2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"6rY1UNA-ZWQM"},"source":["# 1. Basic setup"]},{"cell_type":"code","execution_count":1,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:07.313098989Z","start_time":"2023-12-31T21:19:07.312718292Z"},"executionInfo":{"elapsed":2560,"status":"ok","timestamp":1704531342531,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"CPen6psLZWQM"},"outputs":[],"source":["import os\n","\n","# Set your preferred name\n","USER = \"Konrad\"\n","# Set the data and output directories\n","DATA_ROOT = r'C:\\Users\\conra\\OneDrive\\Desktop\\AI4Science\\DATA'\n","OUTPUT_ROOT = r'C:\\Users\\conra\\OneDrive\\Desktop\\AI4Science\\output1'\n","TRAIT = \"Tourrette Syndrome\"\n","OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n","JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n","if not os.path.exists(OUTPUT_DIR):\n","    os.makedirs(OUTPUT_DIR, exist_ok=True)\n","\n","# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n","NORMALIZE_GENE = True"]},{"cell_type":"code","execution_count":16,"metadata":{},"outputs":[],"source":["import os\n","import mygene\n","import io\n","import re\n","import pandas as pd\n","import gzip\n","import numpy as np\n","from typing import Optional, List, Tuple, Union, Any\n","import json\n","from sparse_lmm import VariableSelection\n","from sklearn.linear_model import Lasso, LogisticRegression\n","from sklearn.metrics import accuracy_score, mean_squared_error\n","from typing import Callable, Optional, List, Tuple, Union, Any\n","from statsmodels.stats.multitest import multipletests\n","\n","def get_relevant_filepaths(cohort_dir):\n","    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n","    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n","    strategies later.\n","    \"\"\"\n","    files = os.listdir(cohort_dir)\n","    soft_files = [f for f in files if 'soft' in f.lower()]\n","    matrix_files = [f for f in files if 'matrix' in f.lower()]\n","    assert len(soft_files) > 0 and len(matrix_files) > 0\n","    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n","    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n","\n","    return soft_file_path, matrix_file_path\n","\n","def line_generator(source, source_type):\n","    \"\"\"Generator that yields lines from a file or a string.\n","\n","    Parameters:\n","    - source: File path or string content.\n","    - source_type: 'file' or 'string'.\n","    \"\"\"\n","    if source_type == 'file':\n","        with gzip.open(source, 'rt') as f:\n","            for line in f:\n","                yield line.strip()\n","    elif source_type == 'string':\n","        for line in source.split('\\n'):\n","            yield line.strip()\n","    else:\n","        raise ValueError(\"source_type must be 'file' or 'string'\")\n","\n","def filter_content_by_prefix(\n","    source: str,\n","    prefixes_a: List[str],\n","    prefixes_b: Optional[List[str]] = None,\n","    unselect: bool = False,\n","    source_type: str = 'file',\n","    return_df_a: bool = True,\n","    return_df_b: bool = True\n",") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n","    \"\"\"\n","    Filters rows from a file or a list of strings based on specified prefixes.\n","\n","    Parameters:\n","    - source (str): File path or string content to filter.\n","    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n","    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n","    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n","    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n","    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n","    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n","\n","    Returns:\n","    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n","    \"\"\"\n","    filtered_lines_a = []\n","    filtered_lines_b = []\n","    prefix_set_a = set(prefixes_a)\n","    if prefixes_b is not None:\n","        prefix_set_b = set(prefixes_b)\n","\n","    # Use generator to get lines\n","    for line in line_generator(source, source_type):\n","        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n","        if matched_a != unselect:\n","            filtered_lines_a.append(line)\n","        if prefixes_b is not None:\n","            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n","            if matched_b != unselect:\n","                filtered_lines_b.append(line)\n","\n","    filtered_content_a = '\\n'.join(filtered_lines_a)\n","    if return_df_a:\n","        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n","    filtered_content_b = None\n","    if filtered_lines_b:\n","        filtered_content_b = '\\n'.join(filtered_lines_b)\n","        if return_df_b:\n","            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n","\n","    return filtered_content_a, filtered_content_b\n","\n","def get_background_and_clinical_data(file_path,\n","                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n","                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n","    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n","    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n","                                                              source_type='file',\n","                                                              return_df_a=False, return_df_b=True)\n","    return background_info, clinical_data\n","\n","def get_unique_values_by_row(dataframe, max_len=30):\n","    \"\"\"\n","    Organize the unique values in each row of the given dataframe, to get a dictionary\n","    :param dataframe:\n","    :param max_len:\n","    :return:\n","    \"\"\"\n","    if '!Sample_geo_accession' in dataframe.columns:\n","        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n","    unique_values_dict = {}\n","    for index, row in dataframe.iterrows():\n","        unique_values = list(row.unique())[:max_len]\n","        unique_values_dict[index] = unique_values\n","    return unique_values_dict\n","\n","def check_rows_and_columns(dataframe, display=False):\n","    \"\"\"\n","    Get the lists of row names and column names of a dataset, and optionally observe them.\n","    :param dataframe:\n","    :param display:\n","    :return:\n","    \"\"\"\n","    dataframe_rows = dataframe.index.tolist()\n","    if display:\n","        print(f\"The dataset has {len(dataframe_rows)} rows, such as {dataframe_rows[:20]}\")\n","    dataframe_cols = dataframe.columns.tolist()\n","    if display:\n","        print(f\"\\nThe dataset has {len(dataframe_cols)} columns, such as {dataframe_cols[:20]}\")\n","    return dataframe_rows, dataframe_cols\n","\n","def preview_df(df, n=5):\n","    return df.head(n).to_dict(orient='list')\n","\n","def xena_convert_trait(row_index: str):\n","    \"\"\"\n","    Convert the trait information from Sample IDs to labels depending on the last two digits.\n","    Tumor types range from 01 - 09, normal types from 10 - 19.\n","    :param row_index: the index value of a row\n","    :return: the converted value\n","    \"\"\"\n","    last_two_digits = int(row_index[-2:])\n","\n","    if 1 <= last_two_digits <= 9:\n","        return 1\n","    elif 10 <= last_two_digits <= 19:\n","        return 0\n","    else:\n","        return -1\n","\n","\n","def xena_convert_gender(cell: str):\n","    \"\"\"Convert the cell content about gender to a binary value\n","    \"\"\"\n","    if isinstance(cell, str):\n","        cell = cell.lower()\n","\n","    if cell == \"female\":\n","        return 0\n","    elif cell == \"male\":\n","        return 1\n","    else:\n","        return None\n","\n","\n","def xena_convert_age(cell: str):\n","    \"\"\"Convert the cell content about age to a numerical value using regular expression\n","    \"\"\"\n","    match = re.search(r'\\d+', str(cell))\n","    if match:\n","        return int(match.group())\n","    else:\n","        return None\n","\n","def xena_select_clinical_features(clinical_df, trait, age_col=None, gender_col=None):\n","    feature_list = []\n","    trait_data = clinical_df.index.to_series().apply(xena_convert_trait).rename(trait)\n","    feature_list.append(trait_data)\n","    if age_col:\n","        age_data = clinical_df[age_col].apply(xena_convert_age).rename(\"Age\")\n","        feature_list.append(age_data)\n","    if gender_col:\n","        gender_data = clinical_df[gender_col].apply(xena_convert_gender).rename(\"Gender\")\n","        feature_list.append(gender_data)\n","    selected_clinical_df = pd.concat(feature_list, axis=1)\n","    return selected_clinical_df\n","\n","def normalize_gene_symbols(gene_symbols, batch_size=1000):\n","    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n","    mg = mygene.MyGeneInfo()\n","    normalized_genes = {}\n","\n","    # Process in batches\n","    for i in range(0, len(gene_symbols), batch_size):\n","        batch = gene_symbols[i:i + batch_size]\n","        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n","\n","        # Update the normalized_genes dictionary with results from this batch\n","        for gene in results:\n","            normalized_genes[gene['query']] = gene.get('symbol', None)\n","\n","    # Return the normalized symbols in the same order as the input\n","    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n","\n","def normalize_gene_symbols_in_index(gene_df):\n","    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n","    Remove the rows where the index failed to be normalized.\"\"\"\n","    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n","    assert len(normalized_gene_list) == len(gene_df.index)\n","    gene_df.index = normalized_gene_list\n","    gene_df = gene_df[gene_df.index.notnull()]\n","    return gene_df\n","\n","def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n","    \"\"\"\n","    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n","    :param dataframe:\n","    :param col_name:\n","    :param min_proportion:\n","    :param min_num:\n","    :return:\n","    \"\"\"\n","    label_counter = dataframe[col_name].value_counts()\n","    total_samples = len(dataframe)\n","    rare_label_num = label_counter.min()\n","    rare_label = label_counter.idxmin()\n","    rare_label_proportion = rare_label_num / total_samples\n","\n","    print(\n","        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n","\n","    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n","    return bool(biased)\n","\n","def judge_continuous_variable_biased(dataframe, col_name):\n","    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n","    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n","    based on quartile statistics combined with its common sense knowledge about this feature.\n","    \"\"\"\n","    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n","    min_value = dataframe[col_name].min()\n","    max_value = dataframe[col_name].max()\n","\n","    # Printing quartile information\n","    print(f\"Quartiles for '{col_name}':\")\n","    print(f\"  25%: {quartiles[0.25]}\")\n","    print(f\"  50% (Median): {quartiles[0.5]}\")\n","    print(f\"  75%: {quartiles[0.75]}\")\n","    print(f\"Min: {min_value}\")\n","    print(f\"Max: {max_value}\")\n","\n","    biased = min_value == max_value\n","\n","    return bool(biased)\n","\n","def judge_and_remove_biased_features(df, trait, trait_type):\n","    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n","    if trait_type == \"binary\":\n","        trait_biased = judge_binary_variable_biased(df, trait)\n","    else:\n","        trait_biased = judge_continuous_variable_biased(df, trait)\n","    if trait_biased:\n","        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n","    else:\n","        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n","    if \"Age\" in df.columns:\n","        age_biased = judge_continuous_variable_biased(df, 'Age')\n","        if age_biased:\n","            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n","            df = df.drop(columns='Age')\n","        else:\n","            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n","    if \"Gender\" in df.columns:\n","        gender_biased = judge_binary_variable_biased(df, 'Gender')\n","        if gender_biased:\n","            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n","            df = df.drop(columns='Gender')\n","        else:\n","            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n","\n","    return trait_biased, df\n","\n","def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n","                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n","    \"\"\"\n","    Add or update information about the usability and quality of a dataset for statistical analysis.\n","\n","    Parameters:\n","    cohort (str): A unique identifier for the dataset.\n","    info_path (str): File path to the JSON file where records are stored.\n","    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n","     preprocessed into a dataframe.\n","    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n","        Required if `is_available` is True.\n","    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n","    note (str, optional): Additional notes about the dataset.\n","\n","    Returns:\n","    None: The function does not return a value but updates or creates a record in the specified JSON file.\n","    \"\"\"\n","    if is_available:\n","        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n","                                                             \"is relevant.\"\n","    is_usable = is_available and (not is_biased)\n","    new_record = {\"is_usable\": is_usable,\n","                  \"is_available\": is_available,\n","                  \"is_biased\": is_biased if is_available else None,\n","                  \"has_age\": \"Age\" in df.columns if is_available else None,\n","                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n","                  \"sample_size\": len(df) if is_available else None,\n","                  \"note\": note}\n","\n","    if not os.path.exists(info_path):\n","        with open(info_path, 'w') as file:\n","            json.dump({}, file)\n","        print(f\"A new JSON file was created at: {info_path}\")\n","\n","    with open(info_path, \"r\") as file:\n","        records = json.load(file)\n","    records[cohort] = new_record\n","\n","    temp_path = info_path + \".tmp\"\n","    try:\n","        with open(temp_path, 'w') as file:\n","            json.dump(records, file)\n","        os.replace(temp_path, info_path)\n","\n","    except Exception as e:\n","        print(f\"An error occurred: {e}\")\n","        if os.path.exists(temp_path):\n","            os.remove(temp_path)\n","        raise\n","\n","def get_feature_data(clinical_df, row_id, feature, convert_fn):\n","    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n","    a binary or continuous variable\"\"\"\n","    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n","    clinical_df.index = [feature]\n","    clinical_df = clinical_df.applymap(convert_fn)\n","\n","    return clinical_df\n","\n","def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n","                                 convert_trait: Callable,\n","                                 age_row: Optional[int] = None,\n","                                 convert_age: Optional[Callable] = None,\n","                                 gender_row: Optional[int] = None,\n","                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n","    \"\"\"\n","    Extracts and processes specific clinical features from a DataFrame representing\n","    sample characteristics in the GEO database series.\n","\n","    Parameters:\n","    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n","    - trait (str): The trait of interest.\n","    - trait_row (int): Row identifier for the trait in the DataFrame.\n","    - convert_trait (Callable): Function to convert trait data into a desired format.\n","    - age_row (int, optional): Row identifier for age data. Default is None.\n","    - convert_age (Callable, optional): Function to convert age data. Default is None.\n","    - gender_row (int, optional): Row identifier for gender data. Default is None.\n","    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n","\n","    Returns:\n","    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n","    \"\"\"\n","    feature_list = []\n","\n","    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n","    feature_list.append(trait_data)\n","    if age_row is not None:\n","        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n","        feature_list.append(age_data)\n","    if gender_row is not None:\n","        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n","        feature_list.append(gender_data)\n","\n","    selected_clinical_df = pd.concat(feature_list, axis=0)\n","    return selected_clinical_df\n","\n","def get_genetic_data(file_path):\n","    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n","    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n","    genetic_data = genetic_data.dropna()\n","    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n","    genetic_data.set_index('ID', inplace=True)\n","\n","    return genetic_data\n","\n","def get_gene_annotation(file_path, prefixes=['^', '!', '#']):\n","    \"\"\"Extract from a SOFT file the gene annotation data\"\"\"\n","    gene_metadata = filter_content_by_prefix(file_path, prefixes_a=prefixes, unselect=True, source_type='file',\n","                                             return_df_a=True)\n","    return gene_metadata[0]\n","\n","def get_gene_mapping(annotation, prob_col, gene_col):\n","    \"\"\"Process the gene annotation to get the mapping between gene names and gene probes.\n","    \"\"\"\n","    mapping_data = annotation.loc[:, [prob_col, gene_col]]\n","    mapping_data = mapping_data.dropna()\n","    mapping_data = mapping_data.rename(columns={gene_col: 'Gene'}).astype({'ID': 'str'})\n","\n","    return mapping_data\n","\n","def apply_gene_mapping(expression_df, mapping_df):\n","    \"\"\"\n","    Converts measured data about gene probes into gene expression data.\n","    Handles the potential many-to-many relationship between probes and genes.\n","\n","    Parameters:\n","    expression_df (DataFrame): A DataFrame with gene expression data, indexed by 'ID'.\n","    mapping_df (DataFrame): A DataFrame mapping 'ID' to 'Gene', with 'ID' as a column.\n","\n","    Returns:\n","    DataFrame: A DataFrame with mean gene expression values, indexed by 'Gene'.\n","    \"\"\"\n","\n","    # Define a regex pattern for splitting gene names\n","    split_pattern = r';|\\|+|/{2,}|,|\\[|\\]|\\(|\\)'\n","\n","    # Split the 'Gene' column in 'mapping_df' using the regex pattern\n","    mapping_df['Gene'] = mapping_df['Gene'].str.split(split_pattern)\n","    mapping_df = mapping_df.explode('Gene')\n","\n","    # Set 'ID' as the index of 'mapping_df' for merging\n","    mapping_df.set_index('ID', inplace=True)\n","\n","    # Merge 'mapping_df' with 'expression_df' by their indices\n","    merged_df = mapping_df.join(expression_df)\n","\n","    # Group by 'Gene' and calculate the mean expression values\n","    gene_expression_df = merged_df.groupby('Gene').mean().dropna()\n","\n","    return gene_expression_df\n","\n","def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n","    \"\"\"\n","    Merge the clinical features and gene expression features from two dataframes into one dataframe\n","    \"\"\"\n","    if 'ID' in genetic_df.columns:\n","        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n","    if 'Gene' in genetic_df.columns:\n","        genetic_df = genetic_df.set_index('Gene')\n","    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n","    return merged_data\n","\n","def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n","    \"\"\"\n","    Reads a JSON file and converts it into a pandas DataFrame.\n","\n","    Args:\n","    json_file (str): The path to the JSON file containing the data.\n","\n","    Returns:\n","    DataFrame: A pandas DataFrame with the JSON data.\n","    \"\"\"\n","    with open(json_file, 'r') as file:\n","        data = json.load(file)\n","    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})\n","\n","def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n","    Union[str, None], pd.DataFrame]:\n","    \"\"\"\n","    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n","\n","    Args:\n","    json_file (str): The path to the JSON file containing the data.\n","    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n","\n","    Returns:\n","    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n","           the filtered and ranked DataFrame.\n","    \"\"\"\n","    # Read the JSON file into a DataFrame\n","    df = read_json_to_dataframe(json_file)\n","\n","    if condition:\n","        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n","    else:\n","        filtered_df = df[df['is_usable'] == True]\n","\n","    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n","    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n","\n","    return best_cohort_id, ranked_df\n","\n","def detect_batch_effect(X):\n","    \"\"\"\n","    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n","\n","    Args:\n","    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n","\n","    Returns:\n","    bool: True if a potential batch effect is detected, False otherwise.\n","    \"\"\"\n","    n_samples = X.shape[0]\n","\n","    # Computing XX^T\n","    XXt = np.dot(X, X.T)\n","\n","    # Compute the eigenvalues of XX^T\n","    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n","    eigen_values = sorted(eigen_values, reverse=True)\n","\n","    # Check for large gaps in the eigenvalues\n","    for i in range(len(eigen_values) - 1):\n","        gap = eigen_values[i] - eigen_values[i + 1]\n","        if gap > 1 / n_samples:  # You may need to adjust this threshold\n","            return True\n","\n","    return False\n","\n","def normalize_data(X_train, X_test=None):\n","    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n","    and optionally the test data\"\"\"\n","    mean = np.mean(X_train, axis=0)\n","    std = np.std(X_train, axis=0)\n","\n","    # Handling columns with std = 0\n","    std_no_zero = np.where(std == 0, 1, std)\n","\n","    # Normalize X_train\n","    X_train_normalized = (X_train - mean) / std_no_zero\n","    # Set normalized values to 0 where std was 0\n","    X_train_normalized[:, std == 0] = 0\n","\n","    if X_test is not None:\n","        X_test_normalized = (X_test - mean) / std_no_zero\n","        X_test_normalized[:, std == 0] = 0\n","    else:\n","        X_test_normalized = None\n","\n","    return X_train_normalized, X_test_normalized\n","\n","\n","class ResidualizationRegressor:\n","    def __init__(self, regression_model_constructor, params=None):\n","        if params is None:\n","            params = {}\n","        self.regression_model = regression_model_constructor(**params)\n","        self.beta_Z = None  # Coefficients for regression of Y on Z\n","        self.beta_X = None  # Coefficients for regression of residual on X\n","        self.neg_log_p_values = None  # Negative logarithm of p-values\n","        self.p_values = None  # Actual p-values\n","\n","    def _reshape_data(self, data):\n","        \"\"\"\n","        Reshape the data to ensure it's in the correct format (2D array).\n","\n","        :param data: The input data (can be 1D or 2D array).\n","        :return: Reshaped 2D array.\n","        \"\"\"\n","        if data.ndim == 1:\n","            return data.reshape(-1, 1)\n","        return data\n","\n","    def _reshape_output(self, data):\n","        \"\"\"\n","        Reshape the output data to ensure it's in the correct format (1D array).\n","\n","        :param data: The output data (can be 1D or 2D array).\n","        :return: Reshaped 1D array.\n","        \"\"\"\n","        if data.ndim == 2 and data.shape[1] == 1:\n","            return data.ravel()\n","        return data\n","\n","    def fit(self, X, Y, Z):\n","        X = self._reshape_data(X)\n","        Y = self._reshape_data(Y)\n","        Z = self._reshape_data(Z)\n","\n","        # Step 1: Linear regression of Y on Z\n","        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n","        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n","        Y_hat = Z_ones @ self.beta_Z\n","        e_Y = Y - Y_hat  # Residual of Y\n","\n","        # Step 2: Regress the residual on X using the included regression model\n","        self.regression_model.fit(X, e_Y)\n","\n","        # Obtain coefficients from the regression model\n","        if hasattr(self.regression_model, 'coef_'):\n","            self.beta_X = self.regression_model.coef_\n","        elif hasattr(self.regression_model, 'getBeta'):\n","            beta_output = self.regression_model.getBeta()\n","            self.beta_X = self._reshape_output(beta_output)\n","\n","        # Obtain negative logarithm of p-values, if available\n","        if hasattr(self.regression_model, 'getNegLogP'):\n","            neg_log_p_output = self.regression_model.getNegLogP()\n","            if neg_log_p_output is not None:\n","                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n","                self.p_values = np.exp(-self.neg_log_p_values)\n","                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n","                p_values_Z = np.full(Z.shape[1], np.nan)\n","                self.p_values = np.concatenate((p_values_Z, self.p_values))\n","\n","    def predict(self, X, Z):\n","        X = self._reshape_data(X)\n","        Z = self._reshape_data(Z)\n","\n","        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n","        ZX = np.column_stack((Z, X))\n","        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n","        return ZX @ combined_beta + self.beta_Z[0]\n","\n","    def get_coefficients(self):\n","        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n","\n","    def get_p_values(self):\n","        return self.p_values\n","\n","    def predict(self, X, Z):\n","        X = self._reshape_data(X)\n","        Z = self._reshape_data(Z)\n","\n","        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n","        ZX = np.column_stack((Z, X))\n","        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n","        return ZX @ combined_beta + self.beta_Z[0]\n","\n","    def get_coefficients(self):\n","        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n","\n","    def get_p_values(self):\n","        return self.p_values\n","\n","\n","def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n","    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n","    indices = np.arange(X.shape[0])\n","    np.random.shuffle(indices)\n","\n","    fold_size = len(X) // k\n","    performances = []\n","\n","    for i in range(k):\n","        # Split data into train and test based on the current fold\n","        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n","        train_indices = np.setdiff1d(indices, test_indices)\n","\n","        X_train, X_test = X[train_indices], X[test_indices]\n","        Y_train, Y_test = Y[train_indices], Y[test_indices]\n","        Z_train, Z_test = Z[train_indices], Z[test_indices]\n","\n","        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n","        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n","\n","        # model = model_constructor(**model_params)\n","        model = ResidualizationRegressor(model_constructor, model_params)\n","        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n","        predictions = model.predict(normalized_X_test, normalized_Z_test)\n","\n","        if target_type == 'binary':\n","            predictions = (predictions > 0.5).astype(int)\n","            Y_test = (Y_test > 0.5).astype(int)\n","            performance = accuracy_score(Y_test, predictions)\n","        elif target_type == 'continuous':\n","            performance = mean_squared_error(Y_test, predictions)\n","\n","        performances.append(performance)\n","\n","    cv_mean = np.mean(performances)\n","    cv_std = np.std(performances)\n","\n","    if target_type == 'binary':\n","        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n","    else:\n","        print(f'The cross-validation MSE is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n","\n","    return cv_mean, cv_std\n","\n","def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n","                     threshold: float = 0.05, save_output: bool = True,\n","                     output_dir: str = './output', model_id: int = 1) -> None:\n","    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n","    consists of one variable about condition and multiple variables about genetic factors.\n","    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n","    p-values or non-zero coefficients, depending on the availability of p-values.\n","\n","    Parameters:\n","    model (Any): The trained regression Model.\n","    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n","    trait (str): The target trait of interest.\n","    condition (str): The specific condition to examine within the model.\n","    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n","    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n","    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n","    model_id (int): The index of the model, 1 or 2.\n","\n","    Returns:\n","    None: This function does not return anything but prints and optionally saves the output.\n","    \"\"\"\n","    coefficients = model.get_coefficients().reshape(-1).tolist()\n","    p_values = model.get_p_values()\n","    if p_values is None:\n","        regression_df = pd.DataFrame({\n","            'Variable': feature_names,\n","            'Coefficient': coefficients\n","        })\n","    else:\n","        regression_df = pd.DataFrame({\n","            'Variable': feature_names,\n","            'Coefficient': coefficients,\n","            'p_value': p_values.reshape(-1).tolist()\n","        })\n","\n","    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n","\n","    print(f\"Effect of the condition on the target variable:\")\n","    print(f\"Variable: {condition}\")\n","    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n","    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n","    if p_values is None:\n","        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n","        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n","        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n","        print(\n","            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n","            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n","    else:\n","        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n","        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n","        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n","        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n","        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n","        print(\n","            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n","            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n","\n","    print(significant_genes_sorted.to_string(index=False))\n","\n","    # Optionally, save this to a CSV file\n","    if save_output:\n","        significant_genes_sorted.to_csv(\n","            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)\n","\n"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"uuyW6piSZWQP"},"source":["# 2. Data preprocessing and selection"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"gpsSPrUeZWQP"},"source":["## 2.1. The TCGA Xena dataset"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"xRKc-zchZWQQ"},"source":["In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."]},{"cell_type":"code","execution_count":23,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.770095084Z","start_time":"2023-12-31T21:19:09.102451553Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4162,"status":"ok","timestamp":1704531366297,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"YPDQJCUYZWQQ","outputId":"a8e16124-e9de-44b9-a9da-0f1866273b33"},"outputs":[{"data":{"text/plain":["['Lower-Grade-Glioma-20240409T095222Z-001.zip',\n"," 'Lower-Grade-Glioma.zip',\n"," 'Tourette-Syndrome.zip']"]},"execution_count":23,"metadata":{},"output_type":"execute_result"}],"source":["import os\n","from utils import *\n","\n","dataset = 'TCGA1'\n","dataset_dir = os.path.join(DATA_ROOT, dataset)\n","os.listdir(dataset_dir)[:10]"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"mTPvkvIcZWQR"},"source":["If no match is found, jump directly to GEO in Part 2.2"]},{"cell_type":"code","execution_count":18,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.918243934Z","start_time":"2023-12-31T21:19:09.909021135Z"},"id":"_7r3vQ4DZWQR"},"outputs":[],"source":["trait_subdir = \"TCGA1_Lower_Grade_Glioma_(LGG)\"\n","cohort = 'Xena'\n","# All the cancer traits in Xena are binary\n","trait_type = 'binary'\n","# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n","is_available = True\n","\n","clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.LGG.sampleMap_LGG_clinicalMatrix')\n","genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.LGG.sampleMap_HiSeqV2_PANCAN.gz')"]},{"cell_type":"code","execution_count":19,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.613198409Z","start_time":"2023-12-31T21:19:10.409688003Z"},"id":"MudwB-_iz7sc"},"outputs":[{"ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: 'C:\\\\Users\\\\conra\\\\OneDrive\\\\Desktop\\\\AI4Science\\\\DATA\\\\TCGA1\\\\TCGA1_Lower_Grade_Glioma_(LGG)\\\\TCGA.LGG.sampleMap_LGG_clinicalMatrix'","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","Cell \u001b[1;32mIn[19], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m clinical_data \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mclinical_data_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m      4\u001b[0m genetic_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(genetic_data_file, compression\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgzip\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, index_col\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[0;32m      5\u001b[0m age_col \u001b[38;5;241m=\u001b[39m gender_col \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\io\\parsers\\readers.py:912\u001b[0m, in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[0;32m    899\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[0;32m    900\u001b[0m     dialect,\n\u001b[0;32m    901\u001b[0m     delimiter,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    908\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[0;32m    909\u001b[0m )\n\u001b[0;32m    910\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[1;32m--> 912\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\io\\parsers\\readers.py:577\u001b[0m, in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    574\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[0;32m    576\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[1;32m--> 577\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    579\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[0;32m    580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\io\\parsers\\readers.py:1407\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[1;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[0;32m   1404\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m   1406\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m-> 1407\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\io\\parsers\\readers.py:1661\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[1;34m(self, f, engine)\u001b[0m\n\u001b[0;32m   1659\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[0;32m   1660\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m-> 1661\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1662\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1663\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1664\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1665\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1668\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1670\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1671\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m   1672\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n","File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\io\\common.py:859\u001b[0m, in \u001b[0;36mget_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[0;32m    854\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m    855\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[0;32m    856\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[0;32m    857\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[0;32m    858\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[1;32m--> 859\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[0;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    862\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    863\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    864\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    865\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    866\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    867\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[0;32m    868\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n","\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'C:\\\\Users\\\\conra\\\\OneDrive\\\\Desktop\\\\AI4Science\\\\DATA\\\\TCGA1\\\\TCGA1_Lower_Grade_Glioma_(LGG)\\\\TCGA.LGG.sampleMap_LGG_clinicalMatrix'"]}],"source":["import pandas as pd\n","\n","clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n","genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n","age_col = gender_col = None"]},{"cell_type":"code","execution_count":20,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.617006167Z","start_time":"2023-12-31T21:19:13.614080653Z"},"id":"At6no_ODZWQR"},"outputs":[{"ename":"NameError","evalue":"name 'clinical_data' is not defined","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)","Cell \u001b[1;32mIn[20], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m _, clinical_data_cols \u001b[38;5;241m=\u001b[39m check_rows_and_columns(\u001b[43mclinical_data\u001b[49m)\n","\u001b[1;31mNameError\u001b[0m: name 'clinical_data' is not defined"]}],"source":["_, clinical_data_cols = check_rows_and_columns(clinical_data)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661538569Z","start_time":"2023-12-31T21:19:13.615946213Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":516,"status":"ok","timestamp":1704460162296,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"IXOwP6DCZWQR","outputId":"f3c0ea1f-2b0a-4c68-faf1-c0641f40bb3e"},"outputs":[{"data":{"text/plain":["['_INTEGRATION',\n"," '_PATIENT',\n"," '_cohort',\n"," '_primary_disease',\n"," '_primary_site',\n"," 'additional_pharmaceutical_therapy',\n"," 'additional_radiation_therapy',\n"," 'additional_surgery_locoregional_procedure',\n"," 'additional_surgery_metastatic_procedure',\n"," 'age_at_initial_pathologic_diagnosis']"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data_cols[:10]"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-28T20:20:11.841042813Z","start_time":"2023-12-28T20:20:11.834458206Z"},"collapsed":false,"id":"i3SiZnBCZWQS"},"source":["Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n","Reference prompt:"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661858700Z","start_time":"2023-12-31T21:19:13.656749487Z"},"colab":{"base_uri":"https://localhost:8080/","height":178},"executionInfo":{"elapsed":482,"status":"ok","timestamp":1704460165840,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"x8P_531oZWQS","outputId":"b99f8e7f-e60c-4240-e834-e793fc9015e7"},"outputs":[{"data":{"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"},"text/plain":["\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_surgery_locoregional_procedure', 'additional_surgery_metastatic_procedure', 'age_at_initial_pathologic_diagnosis', 'animal_insect_allergy_history', 'animal_insect_allergy_types', 'asthma_history', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'days_to_additional_surgery_locoregional_procedure', 'days_to_additional_surgery_metastatic_procedure', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_after_initial_treatment', 'days_to_performance_status_assessment', 'eastern_cancer_oncology_group', 'eczema_history', 'family_history_of_cancer', 'family_history_of_primary_brain_tumor', 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust', 'first_diagnosis_age_of_animal_insect_allergy', 'first_diagnosis_age_of_food_allergy', 'first_presenting_symptom', 'first_presenting_symptom_longest_duration', 'followup_case_report_form_submission_reason', 'followup_treatment_success', 'food_allergy_history', 'food_allergy_types', 'form_completion_date', 'gender', 'hay_fever_history', 'headache_history', 'histological_type', 'history_ionizing_rt_to_head', 'history_of_neoadjuvant_treatment', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'inherited_genetic_syndrome_found', 'inherited_genetic_syndrome_result', 'initial_weight', 'intermediate_dimension', 'is_ffpe', 'karnofsky_performance_score', 'laterality', 'ldh1_mutation_found', 'ldh1_mutation_test_method', 'ldh1_mutation_tested', 'longest_dimension', 'lost_follow_up', 'mental_status_changes', 'mold_or_dust_allergy_history', 'motor_movement_changes', 'neoplasm_histologic_grade', 'new_tumor_event_after_initial_treatment', 'oct_embedded', 'other_dx', 'pathology_report_file_name', 'patient_id', 'performance_status_scale_timing', 'person_neoplasm_cancer_status', 'preoperative_antiseizure_meds', 'preoperative_corticosteroids', 'primary_therapy_outcome_success', 'radiation_therapy', 'sample_type', 'sample_type_id', 'seizure_history', 'sensory_changes', 'shortest_dimension', 'supratentorial_localization', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_location', 'tumor_tissue_site', 'vial_number', 'visual_changes', 'vital_status', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_LGG_mutation', '_GENOMIC_ID_TCGA_LGG_PDMRNAseq', '_GENOMIC_ID_TCGA_LGG_RPPA', '_GENOMIC_ID_TCGA_LGG_mutation_broad_gene', '_GENOMIC_ID_TCGA_LGG_gistic2', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_LGG_miRNA_HiSeq', '_GENOMIC_ID_TCGA_LGG_PDMarrayCNV', '_GENOMIC_ID_data/public/TCGA/LGG/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_LGG_mutation_curated_broad_gene', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2_percentile', '_GENOMIC_ID_TCGA_LGG_hMethyl450_MethylMix', '_GENOMIC_ID_TCGA_LGG_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_LGG_mutation_bcm_gene', '_GENOMIC_ID_TCGA_LGG_hMethyl450', '_GENOMIC_ID_TCGA_LGG_PDMarray', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2', '_GENOMIC_ID_TCGA_LGG_G4502A_07_3', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_LGG_gistic2thd', '_GENOMIC_ID_TCGA_LGG_mutation_ucsc_maf_gene']\\n\""]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["f'''\n","Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n","candidate_age_cols = [col_name1, col_name2, ...]\n","candidate_gender_cols = [col_name1, col_name2, ...]\n","If no columns match a criterion, please provide an empty list.\n","\n","Column names:\n","{clinical_data_cols}\n","'''"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661938833Z","start_time":"2023-12-31T21:19:13.656901848Z"},"id":"2OWLgaibZWQS"},"outputs":[],"source":["candidate_age_cols = ['age_at_initial_pathologic_diagnosis', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust', 'first_diagnosis_age_of_animal_insect_allergy', 'first_diagnosis_age_of_food_allergy', 'year_of_initial_pathologic_diagnosis']\n","\n","candidate_gender_cols = ['gender']"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:57:44.207565572Z","start_time":"2023-12-31T03:57:44.202177544Z"},"collapsed":false,"id":"o29AdGTRZWQS"},"source":["Choose a single column from the candidate columns that record age and gender information respectively.\n","If no column meets the requirement, keep 'age_col' or 'gender_col' to None"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.662078139Z","start_time":"2023-12-31T21:19:13.657006017Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":485,"status":"ok","timestamp":1704460206120,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"ZES_DWC8ZWQS","outputId":"1a397575-0986-4c69-90fb-ae28125ea668"},"outputs":[{"data":{"text/plain":["{'age_at_initial_pathologic_diagnosis': [31.0, 67.0, 44.0, 37.0, 50.0],\n"," 'days_to_birth': [-11509.0, -24578.0, -16297.0, -13565.0, -18494.0],\n"," 'days_to_collection': [nan, nan, nan, nan, nan],\n"," 'days_to_death': [nan, 234.0, 1335.0, 1106.0, nan],\n"," 'days_to_initial_pathologic_diagnosis': [0.0, 0.0, 0.0, 0.0, 0.0],\n"," 'days_to_last_followup': [3574.0, nan, nan, 481.0, 1828.0],\n"," 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust': [nan, nan, nan, nan, nan],\n"," 'first_diagnosis_age_of_animal_insect_allergy': ['12 - 20 Years',\n","  nan,\n","  nan,\n","  nan,\n","  nan],\n"," 'first_diagnosis_age_of_food_allergy': [nan, nan, nan, nan, nan],\n"," 'year_of_initial_pathologic_diagnosis': [2005.0,\n","  2005.0,\n","  2006.0,\n","  2009.0,\n","  2010.0]}"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_age_cols])"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.777159137Z","start_time":"2023-12-31T21:19:13.756210135Z"},"id":"Shy0HJpkZWQS"},"outputs":[],"source":["age_col = 'age_at_initial_pathologic_diagnosis'"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.313425023Z","start_time":"2023-12-31T21:19:14.305221062Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704460210696,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"2dEG2RPBZWQS","outputId":"275a4463-4baa-4c8d-e081-a5e32c598f81"},"outputs":[{"data":{"text/plain":["{'gender': ['FEMALE', 'MALE', 'FEMALE', 'MALE', 'MALE']}"]},"execution_count":15,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_gender_cols])"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.852502962Z","start_time":"2023-12-31T21:19:14.845153121Z"},"id":"kNH4VpYKZWQT"},"outputs":[],"source":["gender_col = 'gender'"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.246065025Z","start_time":"2023-12-31T21:19:15.240279162Z"},"id":"AGQ0X4scZWQT"},"outputs":[],"source":["selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col, gender_col)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.593478652Z","start_time":"2023-12-31T21:19:15.583841656Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":53792,"status":"ok","timestamp":1704460269846,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"sxc-Bd_XZWQT","outputId":"35988f5b-fbbd-43d1-93fc-b477838e183a"},"outputs":[{"name":"stderr","output_type":"stream","text":["INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('GTF2IP1', 2), ('RBMY1A3P', 3), ('HERC2P2', 3), ('WASH3P', 3), ('NUDT9P1', 2), ('SNORD104', 2), ('\n","WARNING:biothings.client:153 input query terms found no hit:\t['C16orf13', 'C16orf11', 'LOC100272146', 'LOC339240', 'NACAP1', 'LOC441204', 'KLRA1', 'FAM183A', 'FA\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:9 input query terms found dup hits:\t[('SUGT1P1', 2), ('SNORD127', 2), ('SNORA62', 4), ('IFITM4P', 8), ('HLA-DRB6', 2), ('FUNDC2P2', 2), \n","WARNING:biothings.client:189 input query terms found no hit:\t['NARFL', 'NFKBIL2', 'LOC150197', 'TMEM84', 'LOC162632', 'PPPDE1', 'PPPDE2', 'C1orf38', 'C1orf31', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:20 input query terms found dup hits:\t[('PIP5K1P1', 2), ('SNORA40', 2), ('SNORD116-24', 2), ('SNORD116-25', 2), ('SNORD116-23', 2), ('SNOR\n","WARNING:biothings.client:149 input query terms found no hit:\t['FAM153C', 'C9orf167', 'CLK2P', 'CCDC76', 'CCDC75', 'CCDC72', 'HIST3H2BB', 'PRAC', 'LOC285780', 'LO\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('SNORD58C', 2), ('UOX', 2), ('UBE2Q2P1', 3), ('PPP4R1L', 2), ('SNORD63', 3), ('SNORD116-22', 2), (\n","WARNING:biothings.client:157 input query terms found no hit:\t['C16orf91', 'LOC100130581', 'SFRS2', 'C16orf5', 'C16orf7', 'C16orf3', 'C21orf7', 'C21orf2', 'LOC389\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('FAM66D', 2), ('FAM66A', 2), ('THSD1P1', 2), ('MSLNL', 2), ('EEF1DP3', 2), ('PGM5P2', 2), ('UBE2MP\n","WARNING:biothings.client:168 input query terms found no hit:\t['LOC284551', 'LOC285548', 'LOC728410', 'LOC541473', 'DULLARD', 'KIAA0368', 'EFTUD1', 'TWISTNB', 'SF\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('S100A7L2', 2), ('POM121L8P', 2), ('SNORD16', 2), ('MEG8', 2), ('RPL23P8', 2), ('KIR3DX1', 5), ('R\n","WARNING:biothings.client:165 input query terms found no hit:\t['TMEM188', 'PDZD3', 'FAM102B', 'FAM102A', 'SMCR7L', 'G6PC', 'OSTCL', 'LOC653544', 'LOC653545', 'USP\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:20 input query terms found dup hits:\t[('SNORD50B', 2), ('PCNAP1', 2), ('SNORA63', 7), ('C5orf60', 2), ('CEACAM22P', 2), ('SNORA16A', 2), \n","WARNING:biothings.client:147 input query terms found no hit:\t['LRRC37A4', 'LOC100131726', 'CPSF3L', 'COL4A3BP', 'PAR1', 'LOC92973', 'MICALCL', 'SMCR7', 'HIST4H4'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('SDHAP3', 2), ('PSORS1C3', 8), ('POM121L10P', 2), ('HLA-J', 9), ('HLA-H', 9), ('HLA-L', 7), ('NME2\n","WARNING:biothings.client:152 input query terms found no hit:\t['LOC100131193', 'KIAA1109', 'DUS2L', 'MYST1', 'IGLL3', 'C12orf69', 'NEURL', 'FAM188B', 'FLJ32063', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('PLGLA', 2), ('PA2G4P4', 2), ('RPL23AP7', 2), ('ST20', 2), ('RPL23AP64', 3), ('DPY19L2P2', 3), ('F\n","WARNING:biothings.client:177 input query terms found no hit:\t['LOC283314', 'RBM9', 'C20orf118', 'LOC150786', 'C20orf112', 'C20orf111', 'C20orf117', 'C20orf114', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('TPTE2P1', 2), ('FOLH1B', 2), ('ID2B', 2), ('MSX2P1', 2), ('RAET1K', 2), ('BMS1P4', 2), ('RPL23AP8\n","WARNING:biothings.client:174 input query terms found no hit:\t['MYEOV2', 'FAM35B2', 'C20orf85', 'FAM175A', 'C17orf73', 'C17orf72', 'C17orf71', 'C17orf70', 'C17orf\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:26 input query terms found dup hits:\t[('CATSPER2P1', 2), ('AQP7P3', 2), ('AQP7P1', 2), ('ATP8B5P', 2), ('OR7E91P', 2), ('ZNF204P', 2), ('\n","WARNING:biothings.client:144 input query terms found no hit:\t['BASE', 'C14orf21', 'C14orf23', 'FAM5B', 'TROVE2', 'C18orf62', 'LOC401052', 'GNB2L1', 'FAM19A4', 'F\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('P2RX6P', 2), ('RP9P', 2), ('SNORD1A', 2), ('SNORD10', 2), ('SNORD19', 2), ('NRADDP', 2), ('AOX2P'\n","WARNING:biothings.client:168 input query terms found no hit:\t['TTC15', 'C4orf29', 'FAM48A', 'C4orf21', 'C3orf65', 'C4orf23', 'C3orf67', 'LOC144742', 'C3orf63', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('SNORA58', 2), ('PMCHL1', 2), ('FAM95B1', 2), ('TMEM191A', 2), ('RPL13P5', 2), ('SBDSP1', 2), ('CY\n","WARNING:biothings.client:143 input query terms found no hit:\t['LOC254312', 'ZNF542', 'CSDAP1', 'FAM73A', 'FAM73B', 'TCTE3', 'ZNF37B', 'LOC728875', 'KIAA1826', 'P\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('ZFP91-CNTF', 2), ('SNORD116-27', 2), ('GSTM2P1', 2), ('C3P1', 2), ('NCF1B', 2), ('SNORD114-4', 2)\n","WARNING:biothings.client:137 input query terms found no hit:\t['LOC100129716', 'C14orf184', 'C14orf181', 'C14orf183', 'C14orf182', 'C7orf26', 'C7orf27', 'C7orf23'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P3', 2), ('ANXA2P2', 2), ('SNORD116-12', 2), ('ZNF286B', 2), ('SAA3P', 2), (\n","WARNING:biothings.client:132 input query terms found no hit:\t['AKR1CL1', 'CCDC21', 'CCDC23', 'TMEM111', 'APOB48R', 'LOC595101', 'KIAA1430', 'KIAA1432', 'ORC4L', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('TCP10L2', 2), ('CXADRP3', 2), ('SNORD116-13', 2), ('TPTE2P3', 2), ('SNORD116-15', 2), ('SNORD116-\n","WARNING:biothings.client:175 input query terms found no hit:\t['DFNB59', 'LOC100170939', 'FAM36A', 'C2orf70', 'C2orf71', 'C2orf77', 'C2orf79', 'LOC100128191', 'HR\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('PMS2CL', 2), ('BRD7P3', 2), ('SCARNA17', 2), ('GABARAPL3', 2), ('MT1IP', 2), ('ADAM6', 3), ('SNOR\n","WARNING:biothings.client:142 input query terms found no hit:\t['C9orf30', 'PEG3AS', 'C9orf37', 'TXNDC3', 'LOC442308', 'FAM46D', 'FAM46A', 'FAM46B', 'FAM46C', 'WDR\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:13 input query terms found dup hits:\t[('FAM197Y2', 3), ('SNORD1C', 2), ('SNORA71B', 2), ('NAT8B', 3), ('SNORA21', 2), ('VENTXP7', 2), ('G\n","WARNING:biothings.client:148 input query terms found no hit:\t['FCGR1B', 'RG9MTD2', 'RARS', 'LOC727677', 'CCRN4L', 'SEPT14', 'HKR1', 'C1orf112', 'LOC100270710', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:13 input query terms found dup hits:\t[('OR6W1P', 3), ('LYPLA2P1', 6), ('RPL23AP53', 2), ('TRPC2', 2), ('RBMY2FP', 2), ('BAGE2', 2), ('POM\n","WARNING:biothings.client:150 input query terms found no hit:\t['SIP1', 'FAM129C', 'GIF', 'TMCO7', 'CXorf61', 'CXorf64', 'ZNF192', 'SEPP1', 'KIRREL', 'MGC16121', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:18 input query terms found dup hits:\t[('ISCA1P1', 2), ('SNORD116-5', 2), ('PCDHB19P', 3), ('NAPSB', 2), ('WASH7P', 2), ('TTTY11', 2), ('H\n","WARNING:biothings.client:135 input query terms found no hit:\t['PPYR1', 'FAM82A2', 'FLJ37453', 'FAM82A1', 'KIAA0947', 'MOBKL2B', 'MOBKL2C', 'MOBKL2A', 'CMAH', 'GR\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-530...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('SNORD12B', 2), ('SNORD116-8', 2), ('SNORD116-4', 2), ('SNORD116-1', 2), ('SNORD116-3', 2), ('SNOR\n","WARNING:biothings.client:78 input query terms found no hit:\t['C14orf174', 'C14orf176', 'C14orf177', 'C14orf179', 'PTPLB', 'SGEF', 'LOC400752', 'GRLF1', 'SPANXE'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"]}],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.074598194Z","start_time":"2023-12-31T21:19:15.973600157Z"},"colab":{"base_uri":"https://localhost:8080/","height":495},"executionInfo":{"elapsed":539,"status":"ok","timestamp":1704460275758,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"52Ks-OVpZWQT","outputId":"be6f55a7-72fc-4a6f-aa62-328358327ced"},"outputs":[{"data":{"text/html":["\n","  <div id=\"df-508b9dd3-d2c2-441d-833e-6546520a6ca3\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Lower Grade Glioma</th>\n","      <th>Age</th>\n","      <th>Gender</th>\n","      <th>ARHGEF10L</th>\n","      <th>HIF3A</th>\n","      <th>RNF17</th>\n","      <th>RNF10</th>\n","      <th>RNF11</th>\n","      <th>RNF13</th>\n","      <th>GTF2IP1</th>\n","      <th>...</th>\n","      <th>SLC7A10</th>\n","      <th>PLA2G2C</th>\n","      <th>TULP2</th>\n","      <th>NPY5R</th>\n","      <th>GNGT2</th>\n","      <th>GNGT1</th>\n","      <th>TULP3</th>\n","      <th>BCL6B</th>\n","      <th>GSTK1</th>\n","      <th>SELP</th>\n","    </tr>\n","    <tr>\n","      <th>sampleID</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>TCGA-CS-4938-01</th>\n","      <td>1</td>\n","      <td>31.0</td>\n","      <td>0.0</td>\n","      <td>1.675408</td>\n","      <td>2.256574</td>\n","      <td>-0.025535</td>\n","      <td>-0.129672</td>\n","      <td>0.512522</td>\n","      <td>1.00569</td>\n","      <td>0.980606</td>\n","      <td>...</td>\n","      <td>2.145814</td>\n","      <td>-0.086682</td>\n","      <td>-0.748878</td>\n","      <td>-0.411617</td>\n","      <td>0.508467</td>\n","      <td>-1.28139</td>\n","      <td>0.587923</td>\n","      <td>-1.931327</td>\n","      <td>-1.200995</td>\n","      <td>-5.525533</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4941-01</th>\n","      <td>1</td>\n","      <td>67.0</td>\n","      <td>1.0</td>\n","      <td>1.677108</td>\n","      <td>1.531874</td>\n","      <td>-0.103335</td>\n","      <td>-0.044972</td>\n","      <td>1.044022</td>\n","      <td>0.68629</td>\n","      <td>0.715906</td>\n","      <td>...</td>\n","      <td>3.563814</td>\n","      <td>-0.086682</td>\n","      <td>2.294922</td>\n","      <td>2.043683</td>\n","      <td>0.703167</td>\n","      <td>-1.28139</td>\n","      <td>-0.211777</td>\n","      <td>-0.691827</td>\n","      <td>-0.067395</td>\n","      <td>-4.274433</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4942-01</th>\n","      <td>1</td>\n","      <td>44.0</td>\n","      <td>0.0</td>\n","      <td>1.451408</td>\n","      <td>0.339574</td>\n","      <td>0.227665</td>\n","      <td>-0.359972</td>\n","      <td>0.479422</td>\n","      <td>0.83969</td>\n","      <td>0.928806</td>\n","      <td>...</td>\n","      <td>2.061014</td>\n","      <td>-0.086682</td>\n","      <td>-0.748878</td>\n","      <td>2.006583</td>\n","      <td>1.287867</td>\n","      <td>-1.28139</td>\n","      <td>-0.521377</td>\n","      <td>-1.996927</td>\n","      <td>-1.020895</td>\n","      <td>-5.096833</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4943-01</th>\n","      <td>1</td>\n","      <td>37.0</td>\n","      <td>1.0</td>\n","      <td>1.159508</td>\n","      <td>1.747574</td>\n","      <td>0.193265</td>\n","      <td>0.049128</td>\n","      <td>0.345022</td>\n","      <td>0.20049</td>\n","      <td>0.736706</td>\n","      <td>...</td>\n","      <td>-0.864186</td>\n","      <td>-0.086682</td>\n","      <td>-0.748878</td>\n","      <td>-0.862817</td>\n","      <td>0.231767</td>\n","      <td>-1.28139</td>\n","      <td>0.760023</td>\n","      <td>-1.260727</td>\n","      <td>-2.100995</td>\n","      <td>-4.541333</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4944-01</th>\n","      <td>1</td>\n","      <td>50.0</td>\n","      <td>1.0</td>\n","      <td>0.847608</td>\n","      <td>3.430774</td>\n","      <td>-0.531035</td>\n","      <td>-0.503272</td>\n","      <td>0.820222</td>\n","      <td>1.21089</td>\n","      <td>0.934706</td>\n","      <td>...</td>\n","      <td>3.457214</td>\n","      <td>-0.086682</td>\n","      <td>-0.159678</td>\n","      <td>1.123883</td>\n","      <td>0.491667</td>\n","      <td>-1.28139</td>\n","      <td>-0.713077</td>\n","      <td>-1.252627</td>\n","      <td>-1.109495</td>\n","      <td>-3.516133</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 17350 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-508b9dd3-d2c2-441d-833e-6546520a6ca3')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-508b9dd3-d2c2-441d-833e-6546520a6ca3 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-508b9dd3-d2c2-441d-833e-6546520a6ca3');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-7157ac91-ac61-4dfc-bd0f-ae2b78f7cec9\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-7157ac91-ac61-4dfc-bd0f-ae2b78f7cec9')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-7157ac91-ac61-4dfc-bd0f-ae2b78f7cec9 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"],"text/plain":["                 Lower Grade Glioma   Age  Gender  ARHGEF10L     HIF3A  \\\n","sampleID                                                                 \n","TCGA-CS-4938-01                   1  31.0     0.0   1.675408  2.256574   \n","TCGA-CS-4941-01                   1  67.0     1.0   1.677108  1.531874   \n","TCGA-CS-4942-01                   1  44.0     0.0   1.451408  0.339574   \n","TCGA-CS-4943-01                   1  37.0     1.0   1.159508  1.747574   \n","TCGA-CS-4944-01                   1  50.0     1.0   0.847608  3.430774   \n","\n","                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n","sampleID                                                          ...   \n","TCGA-CS-4938-01 -0.025535 -0.129672  0.512522  1.00569  0.980606  ...   \n","TCGA-CS-4941-01 -0.103335 -0.044972  1.044022  0.68629  0.715906  ...   \n","TCGA-CS-4942-01  0.227665 -0.359972  0.479422  0.83969  0.928806  ...   \n","TCGA-CS-4943-01  0.193265  0.049128  0.345022  0.20049  0.736706  ...   \n","TCGA-CS-4944-01 -0.531035 -0.503272  0.820222  1.21089  0.934706  ...   \n","\n","                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n","sampleID                                                                     \n","TCGA-CS-4938-01  2.145814 -0.086682 -0.748878 -0.411617  0.508467 -1.28139   \n","TCGA-CS-4941-01  3.563814 -0.086682  2.294922  2.043683  0.703167 -1.28139   \n","TCGA-CS-4942-01  2.061014 -0.086682 -0.748878  2.006583  1.287867 -1.28139   \n","TCGA-CS-4943-01 -0.864186 -0.086682 -0.748878 -0.862817  0.231767 -1.28139   \n","TCGA-CS-4944-01  3.457214 -0.086682 -0.159678  1.123883  0.491667 -1.28139   \n","\n","                    TULP3     BCL6B     GSTK1      SELP  \n","sampleID                                                 \n","TCGA-CS-4938-01  0.587923 -1.931327 -1.200995 -5.525533  \n","TCGA-CS-4941-01 -0.211777 -0.691827 -0.067395 -4.274433  \n","TCGA-CS-4942-01 -0.521377 -1.996927 -1.020895 -5.096833  \n","TCGA-CS-4943-01  0.760023 -1.260727 -2.100995 -4.541333  \n","TCGA-CS-4944-01 -0.713077 -1.252627 -1.109495 -3.516133  \n","\n","[5 rows x 17350 columns]"]},"execution_count":19,"metadata":{},"output_type":"execute_result"}],"source":["merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n","merged_data.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.341121723Z","start_time":"2023-12-31T21:19:16.330544222Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":472,"status":"ok","timestamp":1704460283099,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"LQ-LqBo2ZWQT","outputId":"f8fdabf0-244f-452d-ebcc-9a3e67cf2f5a"},"outputs":[{"name":"stdout","output_type":"stream","text":["The merged dataset contains 529 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:17.222505900Z","start_time":"2023-12-31T21:19:17.218781412Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704460285108,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"aSrJydC2ZWQT","outputId":"8f25b9ee-49a0-48c3-d64a-7e1fea16469a"},"outputs":[{"name":"stdout","output_type":"stream","text":["For the feature 'Lower Grade Glioma', the least common label is '1' with 529 occurrences. This represents 100.00% of the dataset.\n","The distribution of the feature 'Lower Grade Glioma' in this dataset is severely biased.\n","\n","Quartiles for 'Age':\n","  25%: 32.0\n","  50% (Median): 41.0\n","  75%: 53.0\n","Min: 14.0\n","Max: 87.0\n","The distribution of the feature 'Age' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '0.0' with 238 occurrences. This represents 44.99% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"data":{"text/plain":["True"]},"execution_count":21,"metadata":{},"output_type":"execute_result"}],"source":["is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.401536463Z","start_time":"2023-12-31T21:19:17.980683628Z"},"id":"ldageITxZWQT"},"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.405288661Z","start_time":"2023-12-31T21:19:35.402799553Z"},"id":"Kq1zPqtfZWQT"},"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-28T21:28:38.028916303Z","start_time":"2023-12-28T21:28:38.016245426Z"},"collapsed":false,"id":"zNMld4NxZWQT"},"source":["## 2.2. The GEO dataset"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:25:23.253882615Z","start_time":"2023-12-31T03:25:23.244062710Z"},"collapsed":false,"id":"uGxJWgJIZWQT"},"source":["In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."]},{"cell_type":"code","execution_count":24,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.411177575Z","start_time":"2023-12-31T21:19:35.405710636Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2815,"status":"ok","timestamp":1704531373562,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"8-VIPj3fZWQU","outputId":"22c7f08e-2249-4a32-d953-d130a7a7a99c"},"outputs":[{"ename":"FileNotFoundError","evalue":"[WinError 3] The system cannot find the path specified: 'C:\\\\Users\\\\conra\\\\OneDrive\\\\Desktop\\\\AI4Science\\\\DATA\\\\GEO\\\\Tourette-Syndrome'","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","Cell \u001b[1;32mIn[24], line 5\u001b[0m\n\u001b[0;32m      2\u001b[0m trait_subdir \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTourette-Syndrome\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m      4\u001b[0m trait_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(DATA_ROOT, dataset, trait_subdir)\n\u001b[1;32m----> 5\u001b[0m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlistdir\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrait_path\u001b[49m\u001b[43m)\u001b[49m\n","\u001b[1;31mFileNotFoundError\u001b[0m: [WinError 3] The system cannot find the path specified: 'C:\\\\Users\\\\conra\\\\OneDrive\\\\Desktop\\\\AI4Science\\\\DATA\\\\GEO\\\\Tourette-Syndrome'"]}],"source":["dataset = 'GEO'\n","trait_subdir = \"Tourette-Syndrome\"\n","\n","trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n","os.listdir(trait_path)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:39:42.634870142Z","start_time":"2023-12-31T03:39:42.534093295Z"},"collapsed":false,"id":"KJEJBf79ZWQU"},"source":["Repeat the below steps for all the accession numbers"]},{"cell_type":"code","execution_count":13,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.413836016Z","start_time":"2023-12-31T21:19:35.410443976Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1704531600140,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"cCUVGG5KZWQU","outputId":"464a4daf-c5ba-426f-ebc6-3d8f42a60973"},"outputs":[{"data":{"text/plain":["('/content/drive/MyDrive/AI4Science_Public/data/GEO/Tourette-Syndrome/GSE146509/GSE146509_family.soft.gz',\n"," '/content/drive/MyDrive/AI4Science_Public/data/GEO/Tourette-Syndrome/GSE146509/GSE146509_series_matrix.txt.gz')"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":["cohort = accession_num = \"GSE146509\"\n","cohort_dir = os.path.join(trait_path, accession_num)\n","soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n","soft_file, matrix_file"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"Rqk_JLieZWQU"},"source":["### Inital filtering and clinical data preprocessing"]},{"cell_type":"code","execution_count":14,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.492119842Z","start_time":"2023-12-31T21:19:35.413383102Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1629,"status":"ok","timestamp":1704531604685,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"B0jX7hOMZWQU","outputId":"a87d54ec-a5e7-4f16-eb27-1d0f9d254d58"},"outputs":[{"name":"stdout","output_type":"stream","text":["!Series_title\t\"Serum microRNA profiles and functional enrichment analysis in comorbidity between Arnold-Chiari and Tourette syndromes: insights into their related pathological pathways\"\n","!Series_summary\t\"To identify the differential expression of serum microRNA as molecular fingerprints for Arnold-Chiari (AC) and Tourette syndromes (TS), also in comorbid status between both disorders (ACTS).\"\n","!Series_summary\t\"To investigate on differentially expressed miRNA associations with the neuropsychological and neuroimaging parameters for each participant and their potential involvement in AC and TS pathological pathways.\"\n","!Series_overall_design\t\"This study profiled the miRNA expression in sera of 10 AC patients, 11 ACTS patients, 6 TS patients and 8 unaffected controls (NC)\"\n"]}],"source":["background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n","clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n","\n","background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n","print(background_info)"]},{"cell_type":"code","execution_count":15,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.507840422Z","start_time":"2023-12-31T21:19:35.451789107Z"},"colab":{"base_uri":"https://localhost:8080/","height":377},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704531607040,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"_oQdQ_4PZWQU","outputId":"f26fb2f6-e0d9-41fd-d3d4-76ff11b37789"},"outputs":[{"data":{"text/html":["\n","  <div id=\"df-6c2b2223-2f3f-4d09-a19a-29053f26bee1\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>!Sample_geo_accession</th>\n","      <th>GSM4388086</th>\n","      <th>GSM4388087</th>\n","      <th>GSM4388088</th>\n","      <th>GSM4388089</th>\n","      <th>GSM4388090</th>\n","      <th>GSM4388091</th>\n","      <th>GSM4388092</th>\n","      <th>GSM4388093</th>\n","      <th>GSM4388094</th>\n","      <th>...</th>\n","      <th>GSM4388111</th>\n","      <th>GSM4388112</th>\n","      <th>GSM4388113</th>\n","      <th>GSM4388114</th>\n","      <th>GSM4388115</th>\n","      <th>GSM4388116</th>\n","      <th>GSM4388117</th>\n","      <th>GSM4388118</th>\n","      <th>GSM4388119</th>\n","      <th>GSM4388120</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: ACTS</td>\n","      <td>subject status: ACTS</td>\n","      <td>subject status: ACTS</td>\n","      <td>subject status: ACTS</td>\n","      <td>subject status: ACTS</td>\n","      <td>...</td>\n","      <td>subject status: ACTS</td>\n","      <td>subject status: Tourette syndromes (TS)</td>\n","      <td>subject status: Tourette syndromes (TS)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: Arnold-Chiari (AC)</td>\n","      <td>subject status: unaffected controls (NC)</td>\n","      <td>subject status: unaffected controls (NC)</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>age: 14</td>\n","      <td>age: 16</td>\n","      <td>age: 9</td>\n","      <td>age: 14</td>\n","      <td>age: 14</td>\n","      <td>age: 14</td>\n","      <td>age: 9</td>\n","      <td>age: 16</td>\n","      <td>age: 9</td>\n","      <td>...</td>\n","      <td>age: 16</td>\n","      <td>age: 16</td>\n","      <td>age: 17</td>\n","      <td>age: 15</td>\n","      <td>age: 5</td>\n","      <td>age: 15</td>\n","      <td>age: 11</td>\n","      <td>age: 15</td>\n","      <td>age: 13</td>\n","      <td>age: 16</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","      <td>gender: F</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","      <td>gender: F</td>\n","      <td>gender: M</td>\n","      <td>gender: F</td>\n","      <td>...</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","      <td>gender: F</td>\n","      <td>gender: F</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","      <td>gender: F</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","      <td>gender: M</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>...</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","      <td>tissue: serum</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>...</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","      <td>molecule subtype: circulating RNA</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 36 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6c2b2223-2f3f-4d09-a19a-29053f26bee1')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-6c2b2223-2f3f-4d09-a19a-29053f26bee1 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-6c2b2223-2f3f-4d09-a19a-29053f26bee1');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-b0d97fa7-6b7c-48ff-9cfb-998e64b66e6a\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b0d97fa7-6b7c-48ff-9cfb-998e64b66e6a')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-b0d97fa7-6b7c-48ff-9cfb-998e64b66e6a button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"],"text/plain":["         !Sample_geo_accession                          GSM4388086  \\\n","0  !Sample_characteristics_ch1  subject status: Arnold-Chiari (AC)   \n","1  !Sample_characteristics_ch1                             age: 14   \n","2  !Sample_characteristics_ch1                           gender: M   \n","3  !Sample_characteristics_ch1                       tissue: serum   \n","4  !Sample_characteristics_ch1   molecule subtype: circulating RNA   \n","\n","                           GSM4388087                          GSM4388088  \\\n","0  subject status: Arnold-Chiari (AC)  subject status: Arnold-Chiari (AC)   \n","1                             age: 16                              age: 9   \n","2                           gender: M                           gender: F   \n","3                       tissue: serum                       tissue: serum   \n","4   molecule subtype: circulating RNA   molecule subtype: circulating RNA   \n","\n","                           GSM4388089                         GSM4388090  \\\n","0  subject status: Arnold-Chiari (AC)               subject status: ACTS   \n","1                             age: 14                            age: 14   \n","2                           gender: M                          gender: M   \n","3                       tissue: serum                      tissue: serum   \n","4   molecule subtype: circulating RNA  molecule subtype: circulating RNA   \n","\n","                          GSM4388091                         GSM4388092  \\\n","0               subject status: ACTS               subject status: ACTS   \n","1                            age: 14                             age: 9   \n","2                          gender: M                          gender: F   \n","3                      tissue: serum                      tissue: serum   \n","4  molecule subtype: circulating RNA  molecule subtype: circulating RNA   \n","\n","                          GSM4388093                         GSM4388094  ...  \\\n","0               subject status: ACTS               subject status: ACTS  ...   \n","1                            age: 16                             age: 9  ...   \n","2                          gender: M                          gender: F  ...   \n","3                      tissue: serum                      tissue: serum  ...   \n","4  molecule subtype: circulating RNA  molecule subtype: circulating RNA  ...   \n","\n","                          GSM4388111                               GSM4388112  \\\n","0               subject status: ACTS  subject status: Tourette syndromes (TS)   \n","1                            age: 16                                  age: 16   \n","2                          gender: M                                gender: M   \n","3                      tissue: serum                            tissue: serum   \n","4  molecule subtype: circulating RNA        molecule subtype: circulating RNA   \n","\n","                                GSM4388113  \\\n","0  subject status: Tourette syndromes (TS)   \n","1                                  age: 17   \n","2                                gender: F   \n","3                            tissue: serum   \n","4        molecule subtype: circulating RNA   \n","\n","                           GSM4388114                          GSM4388115  \\\n","0  subject status: Arnold-Chiari (AC)  subject status: Arnold-Chiari (AC)   \n","1                             age: 15                              age: 5   \n","2                           gender: F                           gender: M   \n","3                       tissue: serum                       tissue: serum   \n","4   molecule subtype: circulating RNA   molecule subtype: circulating RNA   \n","\n","                           GSM4388116                          GSM4388117  \\\n","0  subject status: Arnold-Chiari (AC)  subject status: Arnold-Chiari (AC)   \n","1                             age: 15                             age: 11   \n","2                           gender: M                           gender: F   \n","3                       tissue: serum                       tissue: serum   \n","4   molecule subtype: circulating RNA   molecule subtype: circulating RNA   \n","\n","                           GSM4388118  \\\n","0  subject status: Arnold-Chiari (AC)   \n","1                             age: 15   \n","2                           gender: M   \n","3                       tissue: serum   \n","4   molecule subtype: circulating RNA   \n","\n","                                 GSM4388119  \\\n","0  subject status: unaffected controls (NC)   \n","1                                   age: 13   \n","2                                 gender: M   \n","3                             tissue: serum   \n","4         molecule subtype: circulating RNA   \n","\n","                                 GSM4388120  \n","0  subject status: unaffected controls (NC)  \n","1                                   age: 16  \n","2                                 gender: M  \n","3                             tissue: serum  \n","4         molecule subtype: circulating RNA  \n","\n","[5 rows x 36 columns]"]},"execution_count":15,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data.head()"]},{"cell_type":"code","execution_count":16,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508015426Z","start_time":"2023-12-31T21:19:35.492746637Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":959,"status":"ok","timestamp":1704531610727,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"jym8RXfxZWQU","outputId":"5a31f9ad-fdb1-47a5-d258-88e53c3de4b1"},"outputs":[{"data":{"text/plain":["{0: ['subject status: Arnold-Chiari (AC)',\n","  'subject status: ACTS',\n","  'subject status: unaffected controls (NC)',\n","  'subject status: Tourette syndromes (TS)'],\n"," 1: ['age: 14',\n","  'age: 16',\n","  'age: 9',\n","  'age: 8',\n","  'age: 11',\n","  'age: 12',\n","  'age: 10',\n","  'age: 17',\n","  'age: 15',\n","  'age: 5',\n","  'age: 13'],\n"," 2: ['gender: M', 'gender: F'],\n"," 3: ['tissue: serum'],\n"," 4: ['molecule subtype: circulating RNA']}"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data_unique = get_unique_values_by_row(clinical_data)\n","clinical_data_unique"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978204446Z","start_time":"2023-12-31T03:58:04.922270095Z"},"collapsed":false,"id":"a9Pe1BKHZWQU"},"source":["Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n","Reference prompt:"]},{"cell_type":"code","execution_count":17,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508232400Z","start_time":"2023-12-31T21:19:35.492903487Z"},"colab":{"base_uri":"https://localhost:8080/","height":178},"executionInfo":{"elapsed":582,"status":"ok","timestamp":1704531615747,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"VO8VaFm6ZWQa","outputId":"8d39e8c5-8395-4819-d635-830f76f7ced5"},"outputs":[{"data":{"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"},"text/plain":["'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Tourette Syndrome\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Tourette Syndrome\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Serum microRNA profiles and functional enrichment analysis in comorbidity between Arnold-Chiari and Tourette syndromes: insights into their related pathological pathways\"\\n!Series_summary\\t\"To identify the differential expression of serum microRNA as molecular fingerprints for Arnold-Chiari (AC) and Tourette syndromes (TS), also in comorbid status between both disorders (ACTS).\"\\n!Series_summary\\t\"To investigate on differentially expressed miRNA associations with the neuropsychological and neuroimaging parameters for each participant and their potential involvement in AC and TS pathological pathways.\"\\n!Series_overall_design\\t\"This study profiled the miRNA expression in sera of 10 AC patients, 11 ACTS patients, 6 TS patients and 8 unaffected controls (NC)\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'subject status: Arnold-Chiari (AC)\\', \\'subject status: ACTS\\', \\'subject status: unaffected controls (NC)\\', \\'subject status: Tourette syndromes (TS)\\'], 1: [\\'age: 14\\', \\'age: 16\\', \\'age: 9\\', \\'age: 8\\', \\'age: 11\\', \\'age: 12\\', \\'age: 10\\', \\'age: 17\\', \\'age: 15\\', \\'age: 5\\', \\'age: 13\\'], 2: [\\'gender: M\\', \\'gender: F\\'], 3: [\\'tissue: serum\\'], 4: [\\'molecule subtype: circulating RNA\\']}\\n'"]},"execution_count":17,"metadata":{},"output_type":"execute_result"}],"source":["f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n","1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n","2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n","   (1) Is there human data available for this trait?\n","   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n","   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n","   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n","\n","Background information about the dataset:\n","{background_info}\n","\n","Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n","{clinical_data_unique}\n","'''"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978240504Z","start_time":"2023-12-31T03:58:04.922365324Z"},"collapsed":false,"id":"00OpftSgZWQb"},"source":["Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n","Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."]},{"cell_type":"code","execution_count":18,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508295660Z","start_time":"2023-12-31T21:19:35.493015671Z"},"executionInfo":{"elapsed":849,"status":"ok","timestamp":1704531661209,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"boW_XbcAZWQb"},"outputs":[],"source":["is_gene_availabe = False\n","trait_row_id = 0\n","age_row_id = 1\n","gender_row_id = 2\n","\n","trait_type = 'binary'"]},{"cell_type":"code","execution_count":19,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508365153Z","start_time":"2023-12-31T21:19:35.493151450Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704531663498,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"oHwwe4oPZWQb","outputId":"131c4984-f44b-41cd-f5b2-e3de379b54c4"},"outputs":[{"name":"stdout","output_type":"stream","text":["This cohort is not usable. Please skip the following steps and jump to the next accession number.\n"]}],"source":["is_available = is_gene_availabe and (trait_row_id is not None)\n","if not is_available:\n","    save_cohort_info(cohort, JSON_PATH, is_available)\n","    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508426069Z","start_time":"2023-12-31T21:19:35.493220101Z"},"id":"uFVuqxemZWQb"},"outputs":[],"source":["# Verify and use the functions generated by GPT\n","\n","def convert_trait(value):\n","    if value == 'primary diagnosis: HEALTHY_CONTROL':\n","        return '0'\n","    elif value == 'primary diagnosis: CROHN_DISEASE':\n","        return '1'\n","    else:\n","        return None\n","\n","\n","\n","def convert_age(age_string):\n","    \"\"\"\n","    Convert age string to a numerical value.\n","    \"\"\"\n","    if age_string.startswith('age: '):\n","        try:\n","            return int(age_string.split('age: ')[1].rstrip('y'))  # Extracting numeric part\n","        except ValueError:\n","            return None  # In case of parsing error\n","    else:\n","        return None  # If age string does not follow expected format\n","\n","\n","\n","def convert_gender(value):\n","    # For 'gender': Binary\n","    if value == 'Sex: Female':\n","        return '1'\n","    elif value == 'Sex: Male':\n","        return '0'\n","    else:\n","        return None  # Unknown gender\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508533013Z","start_time":"2023-12-31T21:19:35.493272992Z"},"colab":{"base_uri":"https://localhost:8080/","height":210},"executionInfo":{"elapsed":351,"status":"ok","timestamp":1704465369824,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"3kQDZKKHZWQb","outputId":"a47252c7-e7b6-4db6-8bd4-7a2d747872b6"},"outputs":[{"data":{"text/html":["\n","  <div id=\"df-adde681f-0b24-4725-9e5b-d0280f9ea94b\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GSM3494884</th>\n","      <th>GSM3494885</th>\n","      <th>GSM3494886</th>\n","      <th>GSM3494887</th>\n","      <th>GSM3494888</th>\n","      <th>GSM3494889</th>\n","      <th>GSM3494890</th>\n","      <th>GSM3494891</th>\n","      <th>GSM3494892</th>\n","      <th>GSM3494893</th>\n","      <th>...</th>\n","      <th>GSM3495040</th>\n","      <th>GSM3495041</th>\n","      <th>GSM3495042</th>\n","      <th>GSM3495043</th>\n","      <th>GSM3495044</th>\n","      <th>GSM3495045</th>\n","      <th>GSM3495046</th>\n","      <th>GSM3495047</th>\n","      <th>GSM3495048</th>\n","      <th>GSM3495049</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>Crohn's Disease</th>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>...</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","      <td>None</td>\n","    </tr>\n","    <tr>\n","      <th>Age</th>\n","      <td>56</td>\n","      <td>None</td>\n","      <td>20</td>\n","      <td>51</td>\n","      <td>37</td>\n","      <td>61</td>\n","      <td>None</td>\n","      <td>31</td>\n","      <td>56</td>\n","      <td>41</td>\n","      <td>...</td>\n","      <td>20</td>\n","      <td>69</td>\n","      <td>69</td>\n","      <td>47</td>\n","      <td>47</td>\n","      <td>56</td>\n","      <td>54</td>\n","      <td>53</td>\n","      <td>50</td>\n","      <td>22</td>\n","    </tr>\n","    <tr>\n","      <th>Gender</th>\n","      <td>0</td>\n","      <td>None</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>None</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>3 rows × 166 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-adde681f-0b24-4725-9e5b-d0280f9ea94b')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-adde681f-0b24-4725-9e5b-d0280f9ea94b button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-adde681f-0b24-4725-9e5b-d0280f9ea94b');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-8d51fcb7-dcdc-4df0-8660-5d080c0b3377\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8d51fcb7-dcdc-4df0-8660-5d080c0b3377')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-8d51fcb7-dcdc-4df0-8660-5d080c0b3377 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"],"text/plain":["                GSM3494884 GSM3494885 GSM3494886 GSM3494887 GSM3494888  \\\n","Crohn's Disease       None       None       None       None       None   \n","Age                     56       None         20         51         37   \n","Gender                   0       None          1          1          0   \n","\n","                GSM3494889 GSM3494890 GSM3494891 GSM3494892 GSM3494893  ...  \\\n","Crohn's Disease       None       None       None       None       None  ...   \n","Age                     61       None         31         56         41  ...   \n","Gender                   0       None          0          1          1  ...   \n","\n","                GSM3495040 GSM3495041 GSM3495042 GSM3495043 GSM3495044  \\\n","Crohn's Disease       None       None       None       None       None   \n","Age                     20         69         69         47         47   \n","Gender                   0          0          1          0          0   \n","\n","                GSM3495045 GSM3495046 GSM3495047 GSM3495048 GSM3495049  \n","Crohn's Disease       None       None       None       None       None  \n","Age                     56         54         53         50         22  \n","Gender                   1          1          1          1          0  \n","\n","[3 rows x 166 columns]"]},"execution_count":137,"metadata":{},"output_type":"execute_result"}],"source":["selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row_id, age_row_id, gender_row_id,\n","                                                      convert_trait, convert_age, convert_gender)\n","selected_clinical_data.head()"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978664193Z","start_time":"2023-12-31T03:58:04.966117261Z"},"collapsed":false,"id":"FYVzuozPZWQb"},"source":["### Genetic data preprocessing and final filtering"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590261560Z","start_time":"2023-12-31T21:19:35.493320362Z"},"colab":{"base_uri":"https://localhost:8080/","height":256},"executionInfo":{"elapsed":1072,"status":"ok","timestamp":1704465424900,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"gVXr4nRrZWQb","outputId":"01a4fab4-ba90-47a5-a392-34a5421a042a"},"outputs":[{"data":{"text/html":["\n","  <div id=\"df-3ef8dedf-dda2-42a4-9759-fe73afd17e84\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ID</th>\n","      <th>GSM3494884</th>\n","      <th>GSM3494885</th>\n","      <th>GSM3494886</th>\n","      <th>GSM3494887</th>\n","      <th>GSM3494888</th>\n","      <th>GSM3494889</th>\n","      <th>GSM3494890</th>\n","      <th>GSM3494891</th>\n","      <th>GSM3494892</th>\n","      <th>...</th>\n","      <th>GSM3495040</th>\n","      <th>GSM3495041</th>\n","      <th>GSM3495042</th>\n","      <th>GSM3495043</th>\n","      <th>GSM3495044</th>\n","      <th>GSM3495045</th>\n","      <th>GSM3495046</th>\n","      <th>GSM3495047</th>\n","      <th>GSM3495048</th>\n","      <th>GSM3495049</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>1</td>\n","      <td>6.948572</td>\n","      <td>6.783931</td>\n","      <td>6.780049</td>\n","      <td>7.159348</td>\n","      <td>7.311038</td>\n","      <td>8.522366</td>\n","      <td>7.208509</td>\n","      <td>7.339519</td>\n","      <td>7.292977</td>\n","      <td>...</td>\n","      <td>6.957283</td>\n","      <td>7.024298</td>\n","      <td>6.953613</td>\n","      <td>6.610053</td>\n","      <td>6.493569</td>\n","      <td>7.024804</td>\n","      <td>7.469483</td>\n","      <td>6.559963</td>\n","      <td>6.866259</td>\n","      <td>7.015812</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>2</td>\n","      <td>4.267766</td>\n","      <td>3.692029</td>\n","      <td>3.649207</td>\n","      <td>4.331090</td>\n","      <td>3.903376</td>\n","      <td>4.191000</td>\n","      <td>3.771017</td>\n","      <td>4.293601</td>\n","      <td>3.963668</td>\n","      <td>...</td>\n","      <td>4.271395</td>\n","      <td>4.137957</td>\n","      <td>4.130696</td>\n","      <td>3.941061</td>\n","      <td>4.238553</td>\n","      <td>3.597590</td>\n","      <td>3.454440</td>\n","      <td>3.558840</td>\n","      <td>4.008234</td>\n","      <td>3.686148</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>3</td>\n","      <td>4.334513</td>\n","      <td>3.981417</td>\n","      <td>3.922257</td>\n","      <td>4.549893</td>\n","      <td>4.140639</td>\n","      <td>4.013236</td>\n","      <td>3.996610</td>\n","      <td>4.933703</td>\n","      <td>3.735861</td>\n","      <td>...</td>\n","      <td>4.303602</td>\n","      <td>4.277528</td>\n","      <td>4.234886</td>\n","      <td>4.187910</td>\n","      <td>4.143072</td>\n","      <td>3.761361</td>\n","      <td>3.791233</td>\n","      <td>3.793128</td>\n","      <td>3.344874</td>\n","      <td>4.103369</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>9</td>\n","      <td>7.140005</td>\n","      <td>7.411580</td>\n","      <td>6.722240</td>\n","      <td>7.117540</td>\n","      <td>6.874786</td>\n","      <td>7.020385</td>\n","      <td>7.031787</td>\n","      <td>6.980904</td>\n","      <td>7.003364</td>\n","      <td>...</td>\n","      <td>7.121634</td>\n","      <td>6.912987</td>\n","      <td>7.270386</td>\n","      <td>7.561971</td>\n","      <td>8.002471</td>\n","      <td>7.195000</td>\n","      <td>7.414168</td>\n","      <td>7.845036</td>\n","      <td>7.263841</td>\n","      <td>7.301888</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>10</td>\n","      <td>4.486670</td>\n","      <td>4.615900</td>\n","      <td>3.966261</td>\n","      <td>4.543677</td>\n","      <td>4.150289</td>\n","      <td>4.216883</td>\n","      <td>4.736946</td>\n","      <td>4.296770</td>\n","      <td>4.462828</td>\n","      <td>...</td>\n","      <td>4.162474</td>\n","      <td>4.563109</td>\n","      <td>3.977201</td>\n","      <td>4.644588</td>\n","      <td>4.734826</td>\n","      <td>4.412295</td>\n","      <td>4.422868</td>\n","      <td>4.763507</td>\n","      <td>5.003758</td>\n","      <td>4.061337</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 167 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3ef8dedf-dda2-42a4-9759-fe73afd17e84')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-3ef8dedf-dda2-42a4-9759-fe73afd17e84 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-3ef8dedf-dda2-42a4-9759-fe73afd17e84');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-43de029e-cb71-42b5-ad54-832d1402ea72\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-43de029e-cb71-42b5-ad54-832d1402ea72')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-43de029e-cb71-42b5-ad54-832d1402ea72 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"],"text/plain":["   ID  GSM3494884  GSM3494885  GSM3494886  GSM3494887  GSM3494888  GSM3494889  \\\n","0   1    6.948572    6.783931    6.780049    7.159348    7.311038    8.522366   \n","1   2    4.267766    3.692029    3.649207    4.331090    3.903376    4.191000   \n","2   3    4.334513    3.981417    3.922257    4.549893    4.140639    4.013236   \n","3   9    7.140005    7.411580    6.722240    7.117540    6.874786    7.020385   \n","4  10    4.486670    4.615900    3.966261    4.543677    4.150289    4.216883   \n","\n","   GSM3494890  GSM3494891  GSM3494892  ...  GSM3495040  GSM3495041  \\\n","0    7.208509    7.339519    7.292977  ...    6.957283    7.024298   \n","1    3.771017    4.293601    3.963668  ...    4.271395    4.137957   \n","2    3.996610    4.933703    3.735861  ...    4.303602    4.277528   \n","3    7.031787    6.980904    7.003364  ...    7.121634    6.912987   \n","4    4.736946    4.296770    4.462828  ...    4.162474    4.563109   \n","\n","   GSM3495042  GSM3495043  GSM3495044  GSM3495045  GSM3495046  GSM3495047  \\\n","0    6.953613    6.610053    6.493569    7.024804    7.469483    6.559963   \n","1    4.130696    3.941061    4.238553    3.597590    3.454440    3.558840   \n","2    4.234886    4.187910    4.143072    3.761361    3.791233    3.793128   \n","3    7.270386    7.561971    8.002471    7.195000    7.414168    7.845036   \n","4    3.977201    4.644588    4.734826    4.412295    4.422868    4.763507   \n","\n","   GSM3495048  GSM3495049  \n","0    6.866259    7.015812  \n","1    4.008234    3.686148  \n","2    3.344874    4.103369  \n","3    7.263841    7.301888  \n","4    5.003758    4.061337  \n","\n","[5 rows x 167 columns]"]},"execution_count":138,"metadata":{},"output_type":"execute_result"}],"source":["genetic_data = get_genetic_data(matrix_file)\n","genetic_data.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590443507Z","start_time":"2023-12-31T21:19:35.536713559Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":365,"status":"ok","timestamp":1704465428938,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"j3WTypxBZWQc","outputId":"09cd06dc-8dc0-493c-a72a-4f1a10eb3e34"},"outputs":[{"data":{"text/plain":["['1',\n"," '2',\n"," '3',\n"," '9',\n"," '10',\n"," '12',\n"," '13',\n"," '14',\n"," '15',\n"," '16',\n"," '18',\n"," '19',\n"," '20',\n"," '21',\n"," '22',\n"," '23',\n"," '24',\n"," '25',\n"," '26',\n"," '27']"]},"execution_count":139,"metadata":{},"output_type":"execute_result"}],"source":["gene_row_ids = genetic_data['ID'][:20].tolist()\n","gene_row_ids"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:30:41.595335164Z","start_time":"2023-12-31T03:30:41.513232329Z"},"collapsed":false,"id":"FrSj5umZZWQc"},"source":["Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"DpeO_Jd3ZWQc"},"source":["Reference prompt:"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590590427Z","start_time":"2023-12-31T21:19:35.536780517Z"},"colab":{"base_uri":"https://localhost:8080/","height":89},"executionInfo":{"elapsed":6,"status":"ok","timestamp":1704465431894,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"kInJXlliZWQc","outputId":"1e3694f7-64c0-4076-fac4-7821bb549e1e"},"outputs":[{"data":{"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"},"text/plain":["\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1', '2', '3', '9', '10', '12', '13', '14', '15', '16', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27']\\n\""]},"execution_count":140,"metadata":{},"output_type":"execute_result"}],"source":["f'''\n","Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n","requires_gene_mapping = (True or False)\n","\n","Row headers:\n","{gene_row_ids}\n","'''"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"wEp4DKmcZWQc"},"source":["If not required, jump directly to the gene normalization step"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590648558Z","start_time":"2023-12-31T21:19:35.536821375Z"},"id":"kv3cD6jRZWQc"},"outputs":[],"source":["requires_gene_mapping = True"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590695457Z","start_time":"2023-12-31T21:19:35.536910836Z"},"id":"TBTqW54jZWQc"},"outputs":[],"source":["if requires_gene_mapping:\n","    gene_annotation = get_gene_annotation(soft_file)[0]\n","    gene_annotation_summary = preview_df(gene_annotation)\n","    gene_annotation_summary"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978931927Z","start_time":"2023-12-31T03:58:04.966328339Z"},"collapsed":false,"id":"Lg3BW_bsZWQc"},"source":["Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n","Reference prompt:"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590742236Z","start_time":"2023-12-31T21:19:35.537050893Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":464,"status":"ok","timestamp":1704465455993,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"kY03dnyLZWQc","outputId":"1a41f9aa-cc70-4b7b-cf91-96ec6062cd62"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {'ID': ['1', '2', '3', '9', '10'], 'ENTREZ_GENE_ID': ['1', '2', '3', '9', '10'], 'SPOT_ID': [1.0, 2.0, 3.0, 9.0, 10.0]}\n","    \n"]}],"source":["if requires_gene_mapping:\n","    print(f'''\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {gene_annotation_summary}\n","    ''')"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590786881Z","start_time":"2023-12-31T21:19:35.537132859Z"},"id":"fD2yNE8vZWQd"},"outputs":[],"source":["if requires_gene_mapping:\n","    probe_id_key = 'ID'\n","    gene_symb_key = 'Gene Symbol'\n","    gene_mapping = get_gene_mapping(gene_annotation, probe_id_key, gene_symb_key)\n","    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590837318Z","start_time":"2023-12-31T21:19:35.537209586Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":46559,"status":"ok","timestamp":1704465070204,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"IYTbjGQJZWQd","outputId":"d0530d9b-53d1-45e6-ae2b-07d900e5098e"},"outputs":[{"name":"stderr","output_type":"stream","text":["INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:33 input query terms found dup hits:\t[('A2MP1', 2), ('AACSP1', 2), ('AADACP1', 2), ('ABCC13', 2), ('ABCC6P1', 2), ('ACTA2-AS1', 2), ('ACT\n","WARNING:biothings.client:64 input query terms found no hit:\t['AAED1', 'AARS', 'AB074162 /// MIR181A2HG', 'ABCC6 /// LOC101930322', 'AC007364.1 /// FLJ32955', 'A\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('ASMTL-AS1', 3), ('ATP13A4-AS1', 2), ('ATP6AP1L', 2), ('ATP8B5P', 2), ('ATXN8OS', 2), ('AZGP1P1', \n","WARNING:biothings.client:242 input query terms found no hit:\t['ARMC4', 'ARNTL', 'ARNTL2', 'ARNTL2-AS1', 'ARSE', 'ASNA1', 'ASUN', 'ATHL1', 'ATP1A3 /// LOC10192713\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('C2orf27A', 2), ('C2orf83', 2), ('C3P1', 2), ('CA5BP1', 2), ('CADM2-AS1', 2), ('CADM3-AS1', 2), ('\n","WARNING:biothings.client:195 input query terms found no hit:\t['C1orf61', 'C1orf64', 'C1orf68', 'C1orf85', 'C1orf86', 'C1orf95', 'C20orf166', 'C20orf166-AS1', 'C2\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('CHIAP2', 2), ('CLCA3P', 2), ('CLEC4GP1', 2), ('CLIP1-AS1', 2), ('CLRN1-AS1', 2), ('CLUHP3', 2), (\n","WARNING:biothings.client:47 input query terms found no hit:\t['CENPBD1', 'CENPBD1P1', 'CEP83-AS1', 'CES1 /// LOC100653057', 'CGNL1 /// LOC101930344 /// LOC101930\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:29 input query terms found dup hits:\t[('CYP2A7P1', 2), ('CYP2B7P', 2), ('CYP4F30P', 2), ('CYP4Z2P', 2), ('DBIL5P', 2), ('DDX11L2', 2), ('\n","WARNING:biothings.client:51 input query terms found no hit:\t['CYP3A7-CYP3AP1', 'CYR61', 'DAB2 /// LOC101926921', 'DAK', 'DAPK1-IT1 /// DAPK1-IT1', 'DARS', 'DDR1\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('EP300-AS1', 2), ('EPB41L4A-AS1', 2), ('ERVH-1', 2), ('ERVK13-1', 2), ('FAHD2CP', 2), ('FAM201A', \n","WARNING:biothings.client:172 input query terms found no hit:\t['EMR1', 'EMR2', 'EMR3', 'ENTHD2', 'EP400NL', 'EPB41L4A-AS2', 'EPRS', 'EPT1', 'ERBB2IP', 'ERO1L', 'E\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:21 input query terms found dup hits:\t[('FOXP1-IT1', 2), ('FRMD8P1', 2), ('FRY-AS1', 2), ('FTH1P5', 2), ('FUT8-AS1', 2), ('GABARAPL3', 2),\n","WARNING:biothings.client:72 input query terms found no hit:\t['FRG1B', 'FTSJ2', 'FUK', 'FYB', 'G6PC', 'GAREM', 'GAREML', 'GARS', 'GATS', 'GATSL2', 'GATSL3', 'GBA\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:30 input query terms found dup hits:\t[('HERC2P7', 3), ('HLA-DRB6', 2), ('HLA-J', 9), ('HMGB3P1', 2), ('HMGN2P46', 2), ('HMMR-AS1', 2), ('\n","WARNING:biothings.client:89 input query terms found no hit:\t['HDGFRP2', 'HDGFRP3', 'HDHD1', 'HEATR2', 'HERC2P3 /// LOC101930458', 'HEXDC', 'HFE2', 'HHLA3', 'HIA\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:45 input query terms found dup hits:\t[('KCNIP2-AS1', 2), ('KIF9-AS1', 2), ('KIR3DX1', 5), ('KLF3-AS1', 2), ('KLHL30-AS1', 2), ('KMT2E-AS1\n","WARNING:biothings.client:116 input query terms found no hit:\t['KAL1', 'KARS', 'KB-431C1.4', 'KCNE1L', 'KCNQ1OT1 /// LOC101927338', 'KDELC1', 'KDELC2', 'KGFLP2', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:76 input query terms found dup hits:\t[('LINC00671', 2), ('LINC00674', 2), ('LINC00685', 4), ('LINC00687', 2), ('LINC00690', 2), ('LINC007\n","WARNING:biothings.client:490 input query terms found no hit:\t['LINC00669', 'LINC00672', 'LINC00675', 'LINC00883', 'LINC00884', 'LINC00889', 'LINC00893', 'LINC008\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:9 input query terms found dup hits:\t[('LPAL2', 2), ('LRP5L', 2), ('LRRC37A4P', 3), ('LRRC37BP1', 2), ('LY6G6E', 8), ('MALAT1', 2), ('MAM\n","WARNING:biothings.client:408 input query terms found no hit:\t['LOC101928700', 'LOC101928703 /// MGC45800', 'LOC101928707', 'LOC101928710', 'LOC101928716', 'LOC10\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:22 input query terms found dup hits:\t[('MCF2L-AS1', 2), ('MED14OS', 2), ('MED15P9', 3), ('MEIS3P1', 2), ('MIATNB', 2), ('MIPEPP3', 2), ('\n","WARNING:biothings.client:56 input query terms found no hit:\t['MB21D1', 'MEF2BNB', 'MESDC1', 'MESDC2', 'METTL10', 'METTL20', 'METTL21B', 'METTL7A', 'METTL7B', 'M\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:29 input query terms found dup hits:\t[('NDUFV2-AS1', 2), ('NEBL-AS1', 2), ('NFYC-AS1', 2), ('NNT-AS1', 2), ('NOP14-AS1', 2), ('NPTN-IT1',\n","WARNING:biothings.client:31 input query terms found no hit:\t['NCR3LG1 /// RP1-239B22.5', 'NCRNA00185 /// TTTY14', 'NDNL2', 'NGFRAP1', 'NHP2L1', 'NOTCH2NL', 'NOV\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('PDCD4-AS1', 2), ('PDZRN3-AS1', 2), ('PLCG1-AS1', 2), ('PMCHL1', 2), ('PMS2P1', 2), ('PMS2P3', 2),\n","WARNING:biothings.client:40 input query terms found no hit:\t['PAXIP1-AS1', 'PAXIP1OS', 'PCDHB18', 'PCDP1', 'PCNX', 'PCNXL2', 'PCNXL3', 'PCNXL4', 'PDDC1', 'PDZD3\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:24 input query terms found dup hits:\t[('PRICKLE2-AS3', 2), ('PRKAG2-AS1', 2), ('PRKG1-AS1', 2), ('PRKXP1', 2), ('PRMT5-AS1', 2), ('PRORY'\n","WARNING:biothings.client:35 input query terms found no hit:\t['PRIM2 /// PRIM2B', 'PRKCDBP', 'PRKRIR', 'PRO2852', 'PROL1', 'PROSC', 'PRR24', 'PRR26', 'PRUNE', 'P\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:26 input query terms found dup hits:\t[('RNF157-AS1', 2), ('RNF185-AS1', 2), ('RP9P', 2), ('RPL23AP32', 2), ('RPL23AP53', 2), ('RPL32P3', \n","WARNING:biothings.client:76 input query terms found no hit:\t['RNF139-AS1', 'RNF144A-AS1', 'RNF165', 'RNF219', 'RNMTL1', 'RP11-203B7.2', 'RP11-215H22.1', 'RP11-2\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('SLC26A4-AS1', 2), ('SMG7-AS1', 2), ('SNHG18', 2), ('SNHG19', 2), ('SNORA21', 2), ('SNORA71A', 2),\n","WARNING:biothings.client:32 input query terms found no hit:\t['SLC26A10', 'SLC2A1-AS1', 'SLC9A3R1', 'SLC9A3R2', 'SLMO1', 'SLMO2', 'SMA4', 'SMC5-AS1', 'SMEK1', 'S\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('STXBP5-AS1', 2), ('SUGT1P1', 2), ('SUGT1P3', 2), ('TAF1A-AS1', 2), ('TAPT1-AS1', 2), ('TEC', 2), \n","WARNING:biothings.client:54 input query terms found no hit:\t['STRA13', 'SUCLG2-AS1', 'SUV420H1', 'SUV420H2', 'SUZ12P /// SUZ12P1', 'T', 'TAAR3', 'TARS', 'TARSL2\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:35 input query terms found dup hits:\t[('TNK2-AS1', 2), ('TP53TG1', 2), ('TPT1-AS1', 2), ('TPT1P8', 2), ('TPTE2P5', 2), ('TPTE2P6', 2), ('\n","WARNING:biothings.client:35 input query terms found no hit:\t['TNRC6C-AS1', 'TOLLIP-AS1', 'TOMM70A', 'TOPORS-AS1', 'TP73-AS1', 'TROVE2', 'TRY2P', 'TSSC1', 'TSTA3\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:23 input query terms found dup hits:\t[('WAC-AS1', 2), ('WFDC21P', 2), ('WHAMMP2', 3), ('WT1-AS', 2), ('YWHAEP7', 3), ('ZBED3-AS1', 2), ('\n","WARNING:biothings.client:49 input query terms found no hit:\t['VWA9', 'WAPAL', 'WARS', 'WASIR1 /// WASIR1 /// WASIR2 /// WASIR2', 'WBP5', 'WBSCR16', 'WBSCR17', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-61...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:2 input query terms found dup hits:\t[('ZNRF2P1', 2), ('ZSCAN12P1', 2)]\n","WARNING:biothings.client:4 input query terms found no hit:\t['ZNRD1', 'ZNRD1-AS1', 'ZUFSP', 'hsa-let-7a-3 /// hsa-let-7b /// hsa-mir-4763 /// MIRLET7BHG /// RP4\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"]}],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590902191Z","start_time":"2023-12-31T21:19:35.537285901Z"},"id":"gGEX1XUUZWQd"},"outputs":[],"source":["merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n","# The preprocessing runs through, which means is_available should be True\n","is_available = True"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.591095170Z","start_time":"2023-12-31T21:19:35.538708445Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704465084232,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"Iy9Qpg0MZWQd","outputId":"dfd783b4-e1c4-4965-a592-6f2da3f0b897"},"outputs":[{"name":"stdout","output_type":"stream","text":["The merged dataset contains 489 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610849182Z","start_time":"2023-12-31T21:19:35.541210258Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":398,"status":"ok","timestamp":1704465086958,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"LXD45pSgZWQd","outputId":"2739b0ea-cb8e-45cd-8812-54e9768870f6"},"outputs":[{"name":"stdout","output_type":"stream","text":["For the feature 'Crohn's Disease', the least common label is '0' with 25 occurrences. This represents 5.11% of the dataset.\n","The distribution of the feature 'Crohn's Disease' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '0' with 243 occurrences. This represents 49.69% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"data":{"text/plain":["False"]},"execution_count":124,"metadata":{},"output_type":"execute_result"}],"source":["is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610923234Z","start_time":"2023-12-31T21:19:35.584848473Z"},"id":"OZ0SAESiZWQd"},"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610970734Z","start_time":"2023-12-31T21:19:35.585042694Z"},"id":"02hAGq1wZWQd"},"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:22.271683755Z","start_time":"2023-12-31T03:58:22.246557674Z"},"id":"-MTPhRGxJV7I"},"source":["### 3. Do regression & Cross Validation"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.459442087Z","start_time":"2023-12-31T21:19:44.416219648Z"},"colab":{"base_uri":"https://localhost:8080/","height":35},"executionInfo":{"elapsed":369,"status":"ok","timestamp":1704465523955,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"2P5jb3pNZWQd","outputId":"3eebf183-ccc5-4c91-a650-3040f88fc417"},"outputs":[{"data":{"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"},"text/plain":["'GSE169568'"]},"execution_count":144,"metadata":{},"output_type":"execute_result"}],"source":["best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH, 'has_age')\n","best_cohort"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.808945960Z","start_time":"2023-12-31T21:19:44.799467072Z"},"colab":{"base_uri":"https://localhost:8080/","height":81},"executionInfo":{"elapsed":371,"status":"ok","timestamp":1704465528445,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"ooo1W8o8ZWQd","outputId":"190a4d3e-025b-487e-edd3-25d70daf996a"},"outputs":[{"data":{"text/html":["\n","  <div id=\"df-fe556a50-0422-4bd2-8107-a3b451273d3f\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>cohort_id</th>\n","      <th>is_usable</th>\n","      <th>is_available</th>\n","      <th>is_biased</th>\n","      <th>has_age</th>\n","      <th>has_gender</th>\n","      <th>sample_size</th>\n","      <th>note</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>1</th>\n","      <td>GSE169568</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>147.0</td>\n","      <td></td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-fe556a50-0422-4bd2-8107-a3b451273d3f')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-fe556a50-0422-4bd2-8107-a3b451273d3f button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-fe556a50-0422-4bd2-8107-a3b451273d3f');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","    </div>\n","  </div>\n"],"text/plain":["   cohort_id  is_usable  is_available is_biased has_age has_gender  \\\n","1  GSE169568       True          True     False    True       True   \n","\n","   sample_size note  \n","1        147.0       "]},"execution_count":145,"metadata":{},"output_type":"execute_result"}],"source":["ranked_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.222719978Z","start_time":"2023-12-31T21:19:45.162873006Z"},"id":"4_XKlkP_ZWQd"},"outputs":[],"source":["merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, best_cohort+'.csv'))"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.234945878Z","start_time":"2023-12-31T21:19:52.231604618Z"},"colab":{"base_uri":"https://localhost:8080/","height":273},"executionInfo":{"elapsed":406,"status":"ok","timestamp":1704465537171,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"vTvOiZMsZWQe","outputId":"59cd4a2f-c12a-4ca2-9578-4d57ddfdd902"},"outputs":[{"data":{"text/html":["\n","  <div id=\"df-66a89291-7aaf-46c8-b58f-645c732c422c\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Crohn's Disease</th>\n","      <th>Age</th>\n","      <th>Gender</th>\n","      <th>A4GNT</th>\n","      <th>AAAS</th>\n","      <th>AACS</th>\n","      <th>AADACL4</th>\n","      <th>AAGAB</th>\n","      <th>AAK1</th>\n","      <th>AAMP</th>\n","      <th>...</th>\n","      <th>ZSWIM7</th>\n","      <th>ZW10</th>\n","      <th>ZWILCH</th>\n","      <th>ZWINT</th>\n","      <th>ZXDB</th>\n","      <th>ZXDC</th>\n","      <th>ZYG11B</th>\n","      <th>ZYX</th>\n","      <th>ZZEF1</th>\n","      <th>ZZZ3</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0</td>\n","      <td>20</td>\n","      <td>1</td>\n","      <td>5.070109</td>\n","      <td>5.559631</td>\n","      <td>6.134333</td>\n","      <td>5.031913</td>\n","      <td>7.645955</td>\n","      <td>8.899405</td>\n","      <td>7.527557</td>\n","      <td>...</td>\n","      <td>6.488454</td>\n","      <td>7.317771</td>\n","      <td>4.979095</td>\n","      <td>4.798338</td>\n","      <td>5.516780</td>\n","      <td>7.573893</td>\n","      <td>10.875069</td>\n","      <td>11.946215</td>\n","      <td>9.423885</td>\n","      <td>7.178631</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0</td>\n","      <td>56</td>\n","      <td>1</td>\n","      <td>5.425123</td>\n","      <td>6.212165</td>\n","      <td>6.170149</td>\n","      <td>5.386187</td>\n","      <td>7.792311</td>\n","      <td>8.849515</td>\n","      <td>8.078419</td>\n","      <td>...</td>\n","      <td>6.256758</td>\n","      <td>7.161959</td>\n","      <td>5.126632</td>\n","      <td>4.644611</td>\n","      <td>5.462659</td>\n","      <td>7.638470</td>\n","      <td>10.438330</td>\n","      <td>12.496480</td>\n","      <td>9.224910</td>\n","      <td>6.157696</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>1</td>\n","      <td>22</td>\n","      <td>0</td>\n","      <td>5.457190</td>\n","      <td>6.376861</td>\n","      <td>6.259700</td>\n","      <td>4.756054</td>\n","      <td>7.343049</td>\n","      <td>8.507237</td>\n","      <td>8.272350</td>\n","      <td>...</td>\n","      <td>7.066220</td>\n","      <td>7.139995</td>\n","      <td>4.736888</td>\n","      <td>4.851393</td>\n","      <td>4.883319</td>\n","      <td>7.697694</td>\n","      <td>10.715323</td>\n","      <td>12.746669</td>\n","      <td>9.165488</td>\n","      <td>5.828003</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0</td>\n","      <td>32</td>\n","      <td>0</td>\n","      <td>5.191521</td>\n","      <td>6.467095</td>\n","      <td>5.896212</td>\n","      <td>5.030326</td>\n","      <td>7.557309</td>\n","      <td>7.677507</td>\n","      <td>7.833617</td>\n","      <td>...</td>\n","      <td>7.312751</td>\n","      <td>6.971038</td>\n","      <td>5.024168</td>\n","      <td>4.761762</td>\n","      <td>4.746781</td>\n","      <td>7.239981</td>\n","      <td>10.512965</td>\n","      <td>12.326666</td>\n","      <td>9.156230</td>\n","      <td>5.973652</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0</td>\n","      <td>32</td>\n","      <td>1</td>\n","      <td>4.993058</td>\n","      <td>6.490008</td>\n","      <td>6.335348</td>\n","      <td>5.092973</td>\n","      <td>7.315477</td>\n","      <td>7.898144</td>\n","      <td>8.096650</td>\n","      <td>...</td>\n","      <td>6.211709</td>\n","      <td>7.728567</td>\n","      <td>4.706430</td>\n","      <td>4.718761</td>\n","      <td>5.738336</td>\n","      <td>7.764994</td>\n","      <td>9.825072</td>\n","      <td>11.528556</td>\n","      <td>9.226991</td>\n","      <td>7.201866</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 9552 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-66a89291-7aaf-46c8-b58f-645c732c422c')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-66a89291-7aaf-46c8-b58f-645c732c422c button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-66a89291-7aaf-46c8-b58f-645c732c422c');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-5157110e-4971-4193-93c4-88c19b187fd6\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5157110e-4971-4193-93c4-88c19b187fd6')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-5157110e-4971-4193-93c4-88c19b187fd6 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"],"text/plain":["   Crohn's Disease  Age  Gender     A4GNT      AAAS      AACS   AADACL4  \\\n","0                0   20       1  5.070109  5.559631  6.134333  5.031913   \n","1                0   56       1  5.425123  6.212165  6.170149  5.386187   \n","2                1   22       0  5.457190  6.376861  6.259700  4.756054   \n","3                0   32       0  5.191521  6.467095  5.896212  5.030326   \n","4                0   32       1  4.993058  6.490008  6.335348  5.092973   \n","\n","      AAGAB      AAK1      AAMP  ...    ZSWIM7      ZW10    ZWILCH     ZWINT  \\\n","0  7.645955  8.899405  7.527557  ...  6.488454  7.317771  4.979095  4.798338   \n","1  7.792311  8.849515  8.078419  ...  6.256758  7.161959  5.126632  4.644611   \n","2  7.343049  8.507237  8.272350  ...  7.066220  7.139995  4.736888  4.851393   \n","3  7.557309  7.677507  7.833617  ...  7.312751  6.971038  5.024168  4.761762   \n","4  7.315477  7.898144  8.096650  ...  6.211709  7.728567  4.706430  4.718761   \n","\n","       ZXDB      ZXDC     ZYG11B        ZYX     ZZEF1      ZZZ3  \n","0  5.516780  7.573893  10.875069  11.946215  9.423885  7.178631  \n","1  5.462659  7.638470  10.438330  12.496480  9.224910  6.157696  \n","2  4.883319  7.697694  10.715323  12.746669  9.165488  5.828003  \n","3  4.746781  7.239981  10.512965  12.326666  9.156230  5.973652  \n","4  5.738336  7.764994   9.825072  11.528556  9.226991  7.201866  \n","\n","[5 rows x 9552 columns]"]},"execution_count":147,"metadata":{},"output_type":"execute_result"}],"source":["merged_data.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.312540150Z","start_time":"2023-12-31T21:19:52.235772334Z"},"id":"M-4hO0PHZWQe"},"outputs":[],"source":["# If both age and gender features are available, select 'age' as the condition.\n","condition = 'Age'\n","# Remove the other condition to prevent interference.\n","merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.323468266Z","start_time":"2023-12-31T21:19:52.290903989Z"},"id":"dhUrhPcqZWQe"},"outputs":[],"source":["X = merged_data.drop(columns=[TRAIT, condition]).values\n","Y = merged_data[TRAIT].values\n","Z = merged_data[condition].values"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"u5Mx3XpAZWQf"},"source":["Select the appropriate regression model depending on whether the dataset shows batch effect."]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.930840782Z","start_time":"2023-12-31T21:19:52.319092572Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":524,"status":"ok","timestamp":1704465549787,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"GdxnyJr4ZWQg","outputId":"7b1fc3a0-4d0e-4298-dd65-48ed3241a4fc"},"outputs":[{"data":{"text/plain":["True"]},"execution_count":150,"metadata":{},"output_type":"execute_result"}],"source":["has_batch_effect = detect_batch_effect(X)\n","has_batch_effect"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.931029722Z","start_time":"2023-12-31T21:20:00.929710828Z"},"id":"uzy8ltp-ZWQg"},"outputs":[],"source":["if has_batch_effect:\n","    model_constructor = VariableSelection\n","    model_params = {}\n","else:\n","    model_constructor = Lasso\n","    model_params = {'alpha': 1.0, 'random_state': 42}"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:20.709459824Z","start_time":"2023-12-31T21:20:00.931657949Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":18120,"status":"ok","timestamp":1704465590597,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"z9tMro7zZWQg","outputId":"af7c0899-341e-4d04-8ac8-979d9aaad239"},"outputs":[{"name":"stdout","output_type":"stream","text":["The cross-validation accuracy is 74.48% ± 3.52%\n"]}],"source":["cv_mean, cv_std = cross_validation(X, Y, Z, model_constructor, model_params, target_type=trait_type)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:24.980870159Z","start_time":"2023-12-31T21:20:20.516710807Z"},"id":"-GzlAnQWZWQg"},"outputs":[],"source":["# Train regression model on the whole dataset to identify significant genes\n","model = ResidualizationRegressor(model_constructor, model_params)\n","normalized_X, _ = normalize_data(X)\n","normalized_Z, _ = normalize_data(Z)\n","model.fit(normalized_X, Y, normalized_Z)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-10-14T04:50:59.072075075Z","start_time":"2023-10-14T04:50:38.739499998Z"},"id":"EjJrxbvb4nlj"},"source":["### 5. Discussion and report"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:25.000830410Z","start_time":"2023-12-31T21:20:24.982834541Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":399,"status":"ok","timestamp":1704465623932,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"FHjsmuPhZWQg","outputId":"01d61e1d-84cd-40da-f37a-b85d3cc91dce"},"outputs":[{"name":"stdout","output_type":"stream","text":["Effect of the condition on the target variable:\n","Variable: Age\n","Coefficient: -0.1141\n","Found 0 significant genes associated with the trait 'Crohn's Disease' conditional on the factor 'Age', with corrected p-value < 0.05:\n","Empty DataFrame\n","Columns: [Variable, Coefficient, corrected_p_value]\n","Index: []\n"]},{"name":"stderr","output_type":"stream","text":["/content/drive/MyDrive/Colab Notebooks/utils.py:425: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"]}],"source":["feature_cols = merged_data.columns.tolist()\n","feature_cols.remove(TRAIT)\n","\n","if has_batch_effect:\n","    report_result_from_lmm(model, feature_cols, TRAIT, condition, threshold=0.05, save_output=True,\n","                           output_dir=OUTPUT_DIR)\n","else:\n","    report_result_from_lasso(model, feature_cols, TRAIT, condition, save_output=True, output_dir=OUTPUT_DIR)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Ia26hwCEZWQg"},"outputs":[],"source":[]}],"metadata":{"colab":{"machine_shape":"hm","provenance":[],"toc_visible":true},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.9"}},"nbformat":4,"nbformat_minor":0}
