{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:02:53.240839272Z",
     "start_time": "2024-01-02T01:02:53.240575109Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "# Set your preferred name\n",
    "USER = \"Jianrong\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/home/yxcheng'\n",
    "OUTPUT_ROOT = '../../output1'\n",
    "TRAIT = 'Fibromyalgia'\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:02:53.501910797Z",
     "start_time": "2024-01-02T01:02:53.496254990Z"
    },
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jMTdsU43vVg3",
    "outputId": "38886111-f442-44b4-8398-96bf384d7abd"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"import os\\nfrom google.colab import drive\\n\\ndrive.mount('/content/drive', force_remount=True)\\nproj_dir = '/content/drive/MyDrive/AI4Science_Public'\\nos.chdir(proj_dir)\""
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n",
    "\n",
    "\"\"\"import os\n",
    "from google.colab import drive\n",
    "\n",
    "drive.mount('/content/drive', force_remount=True)\n",
    "proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n",
    "os.chdir(proj_dir)\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:02:56.462746833Z",
     "start_time": "2024-01-02T01:02:55.818581842Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Bile_Duct_Cancer_(CHOL)',\n",
       " 'TCGA_Lower_Grade_Glioma_(LGG)',\n",
       " 'TCGA_Acute_Myeloid_Leukemia_(LAML)',\n",
       " 'TCGA_Esophageal_Cancer_(ESCA)',\n",
       " 'TCGA_Thyroid_Cancer_(THCA)',\n",
       " 'TCGA_Ocular_melanomas_(UVM)',\n",
       " 'TCGA_Ovarian_Cancer_(OV)',\n",
       " 'TCGA_Prostate_Cancer_(PRAD)',\n",
       " 'CrawlData.ipynb']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "from utils import *\n",
    "\n",
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If no match is found, jump directly to GEO in Part 2.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_subdir = \"TCGA_Breast_Cancer_(BRCA)\"\n",
    "cohort = 'Xena'\n",
    "# All the cancer traits in Xena are binary\n",
    "trait_type = 'binary'\n",
    "# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n",
    "is_available = True\n",
    "\n",
    "clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_BRCA_clinicalMatrix')\n",
    "genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_HiSeqV2_PANCAN.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AJCC_Stage_nature2012',\n",
       " 'Age_at_Initial_Pathologic_Diagnosis_nature2012',\n",
       " 'CN_Clusters_nature2012',\n",
       " 'Converted_Stage_nature2012',\n",
       " 'Days_to_Date_of_Last_Contact_nature2012',\n",
       " 'Days_to_date_of_Death_nature2012',\n",
       " 'ER_Status_nature2012',\n",
       " 'Gender_nature2012',\n",
       " 'HER2_Final_Status_nature2012',\n",
       " 'Integrated_Clusters_no_exp__nature2012']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = ['Age_at_Initial_Pathologic_Diagnosis_nature2012', 'age_at_initial_pathologic_diagnosis',\n",
    "                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = ['Gender_nature2012', 'gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preview_df(clinical_data[candidate_age_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preview_df(clinical_data[candidate_gender_cols])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gender_col = 'gender'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col, gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:02:59.503572381Z",
     "start_time": "2024-01-02T01:02:59.491400342Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE67311']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Fibromyalgia\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:00.287218362Z",
     "start_time": "2024-01-02T01:03:00.278758251Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/home/yxcheng/GEO/Fibromyalgia/GSE67311/GSE67311_family.soft.gz',\n",
       " '/home/yxcheng/GEO/Fibromyalgia/GSE67311/GSE67311_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = accession_num = \"GSE67311\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Inital filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:01.634861460Z",
     "start_time": "2024-01-02T01:03:01.076370517Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Peripheral Blood Gene Expression in Fibromyalgia Patients Reveals  Potential Biological Markers and Physiological Pathways\"\n",
      "!Series_summary\t\"Fibromyalgia (FM) is a common pain disorder characterized by dysregulation in the processing of pain. Although FM has similarities with other rheumatologic pain disorders, the search for objective markers has not been successful. In the current study we analyzed gene expression in the whole blood of 70 fibromyalgia patients and 70 healthy matched controls. Global molecular profiling revealed an upregulation of several inflammatory molecules in FM patients and downregulation of specific pathways related to hypersensitivity and allergy. There was a differential expression of genes in known pathways for pain processing, such as glutamine/glutamate signaling and axonal development. We also identified a panel of candidate gene expression-based classifiers that could establish an objective blood-based molecular diagnostic to objectively identify FM patients and guide design and testing of new therapies. Ten classifier probesets (CPA3, C11orf83, LOC100131943, RGS17, PARD3B, ANKRD20A9P, TTLL7, C8orf12, KAT2B and RIOK3) provided a diagnostic sensitivity of 95% and a specificity of 96%. Molecular scores developed from these classifiers were able to clearly distinguish FM patients from healthy controls. An understanding of molecular dysregulation in fibromyalgia is in its infancy; however the results described herein indicate blood global gene expression profiling provides many testable hypotheses that deserve further exploration.\"\n",
      "!Series_overall_design\t\"Blood samples were collected in PAXgene tubes and collected samples were stored at -80oC.  The RNA was isolated using the PAXgene RNA isolation kit according to standard protocols. Total RNA was quantified on a Nanodrop spectrophotometer and visualized for quality on an Agilent Bioanalyzer. Samples with an average RIN (RNA Integrity Number) >8, indicating good quality RNA, were processed. 200ng of total RNA was amplified and then hybridized to Affymetrix® Human Gene 1.1 ST Peg arrays using standard manufacturer’s protocols on a Gene Titan MC instrument. Data was analyzed using Partek Genomics Suite (version 6.6) using RMA normalization. All genes with Log2 signal intensity less than 4.8 were excluded from analysis based on low expression. Differential expression analysis was carried out using a one way ANOVA by using Method of Moments and Fisher's Least Significant Difference (LSD) tests with a threshold p-value <0.005 for the biological and molecular function analyses, and a more conservative threshold p-value <0.001 (FDR q-value range 0.002% to 13%) for candidate diagnostic signatures.\"\n"
     ]
    }
   ],
   "source": [
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:01.642310454Z",
     "start_time": "2024-01-02T01:03:01.635048940Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1644447</th>\n",
       "      <th>GSM1644448</th>\n",
       "      <th>GSM1644449</th>\n",
       "      <th>GSM1644450</th>\n",
       "      <th>GSM1644451</th>\n",
       "      <th>GSM1644452</th>\n",
       "      <th>GSM1644453</th>\n",
       "      <th>GSM1644454</th>\n",
       "      <th>GSM1644455</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1644579</th>\n",
       "      <th>GSM1644580</th>\n",
       "      <th>GSM1644581</th>\n",
       "      <th>GSM1644582</th>\n",
       "      <th>GSM1644583</th>\n",
       "      <th>GSM1644584</th>\n",
       "      <th>GSM1644585</th>\n",
       "      <th>GSM1644586</th>\n",
       "      <th>GSM1644587</th>\n",
       "      <th>GSM1644588</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>...</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>fiqr score: 8.5</td>\n",
       "      <td>fiqr score: -2.0</td>\n",
       "      <td>fiqr score: 9.8</td>\n",
       "      <td>fiqr score: 0.5</td>\n",
       "      <td>fiqr score: -1.0</td>\n",
       "      <td>fiqr score: -0.5</td>\n",
       "      <td>fiqr score: 2.2</td>\n",
       "      <td>fiqr score: -2.0</td>\n",
       "      <td>fiqr score: -2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>fiqr score: 41.8</td>\n",
       "      <td>fiqr score: 54.5</td>\n",
       "      <td>fiqr score: 63.0</td>\n",
       "      <td>fiqr score: 64.0</td>\n",
       "      <td>fiqr score: 67.2</td>\n",
       "      <td>fiqr score: 17.8</td>\n",
       "      <td>fiqr score: 17.8</td>\n",
       "      <td>fiqr score: 41.7</td>\n",
       "      <td>fiqr score: 81.2</td>\n",
       "      <td>fiqr score: 54.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>bmi: 36</td>\n",
       "      <td>bmi: 34</td>\n",
       "      <td>bmi: 33</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 28</td>\n",
       "      <td>bmi: 23</td>\n",
       "      <td>bmi: 48</td>\n",
       "      <td>bmi: 48</td>\n",
       "      <td>...</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 25</td>\n",
       "      <td>bmi: 29</td>\n",
       "      <td>bmi: 38</td>\n",
       "      <td>bmi: 28</td>\n",
       "      <td>bmi: 37</td>\n",
       "      <td>bmi: 0</td>\n",
       "      <td>bmi: 38</td>\n",
       "      <td>bmi: 20</td>\n",
       "      <td>bmi: 34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>...</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: -</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 143 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                  GSM1644447  \\\n",
       "0  !Sample_characteristics_ch1  diagnosis: healthy control   \n",
       "1  !Sample_characteristics_ch1    tissue: peripheral blood   \n",
       "2  !Sample_characteristics_ch1             fiqr score: 8.5   \n",
       "3  !Sample_characteristics_ch1                     bmi: 36   \n",
       "4  !Sample_characteristics_ch1                migraine: No   \n",
       "\n",
       "                   GSM1644448                  GSM1644449  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood   \n",
       "2            fiqr score: -2.0             fiqr score: 9.8   \n",
       "3                     bmi: 34                     bmi: 33   \n",
       "4                migraine: No                migraine: No   \n",
       "\n",
       "                   GSM1644450                  GSM1644451  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood   \n",
       "2             fiqr score: 0.5            fiqr score: -1.0   \n",
       "3                     bmi: 22                     bmi: 24   \n",
       "4                migraine: No                migraine: No   \n",
       "\n",
       "                   GSM1644452                  GSM1644453  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood   \n",
       "2            fiqr score: -0.5             fiqr score: 2.2   \n",
       "3                     bmi: 28                     bmi: 23   \n",
       "4                migraine: No                migraine: No   \n",
       "\n",
       "                   GSM1644454                  GSM1644455  ...  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control  ...   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood  ...   \n",
       "2            fiqr score: -2.0            fiqr score: -2.0  ...   \n",
       "3                     bmi: 48                     bmi: 48  ...   \n",
       "4                migraine: No                migraine: No  ...   \n",
       "\n",
       "                 GSM1644579                GSM1644580  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 41.8          fiqr score: 54.5   \n",
       "3                   bmi: 22                   bmi: 25   \n",
       "4              migraine: No             migraine: Yes   \n",
       "\n",
       "                 GSM1644581                GSM1644582  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 63.0          fiqr score: 64.0   \n",
       "3                   bmi: 29                   bmi: 38   \n",
       "4             migraine: Yes              migraine: No   \n",
       "\n",
       "                 GSM1644583                GSM1644584  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 67.2          fiqr score: 17.8   \n",
       "3                   bmi: 28                   bmi: 37   \n",
       "4             migraine: Yes             migraine: Yes   \n",
       "\n",
       "                 GSM1644585                GSM1644586  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 17.8          fiqr score: 41.7   \n",
       "3                    bmi: 0                   bmi: 38   \n",
       "4               migraine: -             migraine: Yes   \n",
       "\n",
       "                 GSM1644587                GSM1644588  \n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia  \n",
       "1  tissue: peripheral blood  tissue: peripheral blood  \n",
       "2          fiqr score: 81.2          fiqr score: 54.7  \n",
       "3                   bmi: 20                   bmi: 34  \n",
       "4             migraine: Yes              migraine: No  \n",
       "\n",
       "[5 rows x 143 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:01.805026160Z",
     "start_time": "2024-01-02T01:03:01.804428154Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['diagnosis: healthy control', 'diagnosis: fibromyalgia'],\n",
       " 1: ['tissue: peripheral blood'],\n",
       " 2: ['fiqr score: 8.5',\n",
       "  'fiqr score: -2.0',\n",
       "  'fiqr score: 9.8',\n",
       "  'fiqr score: 0.5',\n",
       "  'fiqr score: -1.0',\n",
       "  'fiqr score: -0.5',\n",
       "  'fiqr score: 2.2',\n",
       "  'fiqr score: 15.3',\n",
       "  'fiqr score: 4.0',\n",
       "  'fiqr score: 29.3',\n",
       "  'fiqr score: 27.2',\n",
       "  'fiqr score: 5.0',\n",
       "  'fiqr score: 1.0',\n",
       "  'fiqr score: 2.5',\n",
       "  'fiqr score: 3.0',\n",
       "  'fiqr score: -1.5',\n",
       "  'fiqr score: 1.3',\n",
       "  'fiqr score: 21.7',\n",
       "  'fiqr score: -1.2',\n",
       "  'fiqr score: 4.3',\n",
       "  'fiqr score: 6.5',\n",
       "  'fiqr score: 2.0',\n",
       "  'fiqr score: 11.7',\n",
       "  'fiqr score: 15.0',\n",
       "  'fiqr score: 6.0',\n",
       "  'fiqr score: 14.2',\n",
       "  'fiqr score: -0.2',\n",
       "  'fiqr score: 12.8',\n",
       "  'fiqr score: 15.7',\n",
       "  'fiqr score: 0.0'],\n",
       " 3: ['bmi: 36',\n",
       "  'bmi: 34',\n",
       "  'bmi: 33',\n",
       "  'bmi: 22',\n",
       "  'bmi: 24',\n",
       "  'bmi: 28',\n",
       "  'bmi: 23',\n",
       "  'bmi: 48',\n",
       "  'bmi: 25',\n",
       "  'bmi: 46',\n",
       "  'bmi: 32',\n",
       "  'bmi: 31',\n",
       "  'bmi: 21',\n",
       "  'bmi: 27',\n",
       "  'bmi: 39',\n",
       "  'bmi: 52',\n",
       "  'bmi: 37',\n",
       "  'bmi: 0',\n",
       "  'bmi: 38',\n",
       "  'bmi: 26',\n",
       "  'bmi: 42',\n",
       "  'bmi: 20',\n",
       "  'bmi: 30',\n",
       "  'bmi: 43',\n",
       "  'bmi: 35',\n",
       "  'bmi: 44',\n",
       "  'bmi: 29',\n",
       "  'bmi: 45',\n",
       "  'bmi: 40',\n",
       "  'bmi: 47'],\n",
       " 4: ['migraine: No', 'migraine: Yes', 'migraine: -'],\n",
       " 5: ['irritable bowel syndrome: No',\n",
       "  'irritable bowel syndrome: Yes',\n",
       "  'irritable bowel syndrome: -'],\n",
       " 6: ['major depression: No', 'major depression: -', 'major depression: Yes'],\n",
       " 7: ['bipolar disorder: No', 'bipolar disorder: -', 'bipolar disorder: Yes'],\n",
       " 8: ['chronic fatigue syndrome: No',\n",
       "  nan,\n",
       "  'chronic fatigue syndrome: -',\n",
       "  'chronic fatigue syndrome: Yes']}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:02.369775044Z",
     "start_time": "2024-01-02T01:03:02.356501629Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Fibromyalgia\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Fibromyalgia\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Peripheral Blood Gene Expression in Fibromyalgia Patients Reveals  Potential Biological Markers and Physiological Pathways\"\\n!Series_summary\\t\"Fibromyalgia (FM) is a common pain disorder characterized by dysregulation in the processing of pain. Although FM has similarities with other rheumatologic pain disorders, the search for objective markers has not been successful. In the current study we analyzed gene expression in the whole blood of 70 fibromyalgia patients and 70 healthy matched controls. Global molecular profiling revealed an upregulation of several inflammatory molecules in FM patients and downregulation of specific pathways related to hypersensitivity and allergy. There was a differential expression of genes in known pathways for pain processing, such as glutamine/glutamate signaling and axonal development. We also identified a panel of candidate gene expression-based classifiers that could establish an objective blood-based molecular diagnostic to objectively identify FM patients and guide design and testing of new therapies. Ten classifier probesets (CPA3, C11orf83, LOC100131943, RGS17, PARD3B, ANKRD20A9P, TTLL7, C8orf12, KAT2B and RIOK3) provided a diagnostic sensitivity of 95% and a specificity of 96%. Molecular scores developed from these classifiers were able to clearly distinguish FM patients from healthy controls. An understanding of molecular dysregulation in fibromyalgia is in its infancy; however the results described herein indicate blood global gene expression profiling provides many testable hypotheses that deserve further exploration.\"\\n!Series_overall_design\\t\"Blood samples were collected in PAXgene tubes and collected samples were stored at -80oC.  The RNA was isolated using the PAXgene RNA isolation kit according to standard protocols. Total RNA was quantified on a Nanodrop spectrophotometer and visualized for quality on an Agilent Bioanalyzer. Samples with an average RIN (RNA Integrity Number) >8, indicating good quality RNA, were processed. 200ng of total RNA was amplified and then hybridized to Affymetrix® Human Gene 1.1 ST Peg arrays using standard manufacturer’s protocols on a Gene Titan MC instrument. Data was analyzed using Partek Genomics Suite (version 6.6) using RMA normalization. All genes with Log2 signal intensity less than 4.8 were excluded from analysis based on low expression. Differential expression analysis was carried out using a one way ANOVA by using Method of Moments and Fisher\\'s Least Significant Difference (LSD) tests with a threshold p-value <0.005 for the biological and molecular function analyses, and a more conservative threshold p-value <0.001 (FDR q-value range 0.002% to 13%) for candidate diagnostic signatures.\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'diagnosis: healthy control\\', \\'diagnosis: fibromyalgia\\'], 1: [\\'tissue: peripheral blood\\'], 2: [\\'fiqr score: 8.5\\', \\'fiqr score: -2.0\\', \\'fiqr score: 9.8\\', \\'fiqr score: 0.5\\', \\'fiqr score: -1.0\\', \\'fiqr score: -0.5\\', \\'fiqr score: 2.2\\', \\'fiqr score: 15.3\\', \\'fiqr score: 4.0\\', \\'fiqr score: 29.3\\', \\'fiqr score: 27.2\\', \\'fiqr score: 5.0\\', \\'fiqr score: 1.0\\', \\'fiqr score: 2.5\\', \\'fiqr score: 3.0\\', \\'fiqr score: -1.5\\', \\'fiqr score: 1.3\\', \\'fiqr score: 21.7\\', \\'fiqr score: -1.2\\', \\'fiqr score: 4.3\\', \\'fiqr score: 6.5\\', \\'fiqr score: 2.0\\', \\'fiqr score: 11.7\\', \\'fiqr score: 15.0\\', \\'fiqr score: 6.0\\', \\'fiqr score: 14.2\\', \\'fiqr score: -0.2\\', \\'fiqr score: 12.8\\', \\'fiqr score: 15.7\\', \\'fiqr score: 0.0\\'], 3: [\\'bmi: 36\\', \\'bmi: 34\\', \\'bmi: 33\\', \\'bmi: 22\\', \\'bmi: 24\\', \\'bmi: 28\\', \\'bmi: 23\\', \\'bmi: 48\\', \\'bmi: 25\\', \\'bmi: 46\\', \\'bmi: 32\\', \\'bmi: 31\\', \\'bmi: 21\\', \\'bmi: 27\\', \\'bmi: 39\\', \\'bmi: 52\\', \\'bmi: 37\\', \\'bmi: 0\\', \\'bmi: 38\\', \\'bmi: 26\\', \\'bmi: 42\\', \\'bmi: 20\\', \\'bmi: 30\\', \\'bmi: 43\\', \\'bmi: 35\\', \\'bmi: 44\\', \\'bmi: 29\\', \\'bmi: 45\\', \\'bmi: 40\\', \\'bmi: 47\\'], 4: [\\'migraine: No\\', \\'migraine: Yes\\', \\'migraine: -\\'], 5: [\\'irritable bowel syndrome: No\\', \\'irritable bowel syndrome: Yes\\', \\'irritable bowel syndrome: -\\'], 6: [\\'major depression: No\\', \\'major depression: -\\', \\'major depression: Yes\\'], 7: [\\'bipolar disorder: No\\', \\'bipolar disorder: -\\', \\'bipolar disorder: Yes\\'], 8: [\\'chronic fatigue syndrome: No\\', nan, \\'chronic fatigue syndrome: -\\', \\'chronic fatigue syndrome: Yes\\']}\\n'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:03.079540570Z",
     "start_time": "2024-01-02T01:03:03.073638535Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = 2\n",
    "gender_row = 3\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:03.572682163Z",
     "start_time": "2024-01-02T01:03:03.559775513Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:03.973964287Z",
     "start_time": "2024-01-02T01:03:03.964958096Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 0.2, 1)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Brain':\n",
    "        return 1  # Epilepsy present\n",
    "    else:\n",
    "        return 0  # Epilepsy not present\n",
    "\n",
    "def convert_age(age_string):\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    if age_string.startswith('age: '): \n",
    "        try:\n",
    "            \n",
    "            age_value = float(age_string.split('age: ')[1].rstrip(' weeks'))\n",
    "            return age_value\n",
    "        except ValueError:\n",
    "            return None \n",
    "    else:\n",
    "        return None  \n",
    "\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary representation.\n",
    "    Female -> 0, Male -> 1, Unknown -> None\n",
    "    \"\"\"\n",
    "    if gender_string == 'gender: female':\n",
    "        return 0\n",
    "    elif gender_string == 'gender: male':\n",
    "        return 1\n",
    "    else:\n",
    "        return None  # In case of unknown gender\n",
    "\n",
    "# Example usage\n",
    "example_tissue = 'tissue: Heart'\n",
    "example_age = 'age: 0.2 weeks'\n",
    "example_gender = 'gender: male'\n",
    "\n",
    "converted_trait = convert_trait(example_tissue)\n",
    "converted_age = convert_age(example_age)\n",
    "converted_gender = convert_gender(example_gender)\n",
    "\n",
    "converted_trait, converted_age, converted_gender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:04.637878352Z",
     "start_time": "2024-01-02T01:03:04.593886885Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yxcheng/AI4Science_gold_standard/stage1/Yuxuan/../utils.py:208: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "/home/yxcheng/AI4Science_gold_standard/stage1/Yuxuan/../utils.py:208: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n",
      "/home/yxcheng/AI4Science_gold_standard/stage1/Yuxuan/../utils.py:208: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM4064970</th>\n",
       "      <th>GSM4064971</th>\n",
       "      <th>GSM4064972</th>\n",
       "      <th>GSM4064973</th>\n",
       "      <th>GSM4064974</th>\n",
       "      <th>GSM4064975</th>\n",
       "      <th>GSM4064976</th>\n",
       "      <th>GSM4064977</th>\n",
       "      <th>GSM4064978</th>\n",
       "      <th>GSM4064979</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4065020</th>\n",
       "      <th>GSM4065021</th>\n",
       "      <th>GSM4065022</th>\n",
       "      <th>GSM4065023</th>\n",
       "      <th>GSM4065024</th>\n",
       "      <th>GSM4065025</th>\n",
       "      <th>GSM4065026</th>\n",
       "      <th>GSM4065027</th>\n",
       "      <th>GSM4065028</th>\n",
       "      <th>GSM4065029</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Brugada Syndrome</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <td>24.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>72.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>60.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GSM4064970  GSM4064971  GSM4064972  GSM4064973  GSM4064974  \\\n",
       "Brugada Syndrome         0.0         0.0         0.0         0.0         0.0   \n",
       "Age                     24.0       112.0         8.0        24.0         0.6   \n",
       "Gender                   1.0         1.0         0.0         1.0         0.0   \n",
       "\n",
       "                  GSM4064975  GSM4064976  GSM4064977  GSM4064978  GSM4064979  \\\n",
       "Brugada Syndrome         0.0         0.0         0.0         0.0         0.0   \n",
       "Age                     72.0        24.0        36.0        52.0        20.0   \n",
       "Gender                   1.0         0.0         1.0         1.0         1.0   \n",
       "\n",
       "                  ...  GSM4065020  GSM4065021  GSM4065022  GSM4065023  \\\n",
       "Brugada Syndrome  ...         1.0         1.0         1.0         1.0   \n",
       "Age               ...        60.0        52.0         0.0         0.0   \n",
       "Gender            ...         0.0         0.0         0.0         0.0   \n",
       "\n",
       "                  GSM4065024  GSM4065025  GSM4065026  GSM4065027  GSM4065028  \\\n",
       "Brugada Syndrome         1.0         1.0         1.0         1.0         1.0   \n",
       "Age                     52.0         0.0         0.0         2.0         2.0   \n",
       "Gender                   0.0         1.0         0.0         1.0         0.0   \n",
       "\n",
       "                  GSM4065029  \n",
       "Brugada Syndrome         1.0  \n",
       "Age                    144.0  \n",
       "Gender                   1.0  \n",
       "\n",
       "[3 rows x 60 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:06.246474256Z",
     "start_time": "2024-01-02T01:03:05.670354792Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM4064970</th>\n",
       "      <th>GSM4064971</th>\n",
       "      <th>GSM4064972</th>\n",
       "      <th>GSM4064973</th>\n",
       "      <th>GSM4064974</th>\n",
       "      <th>GSM4064975</th>\n",
       "      <th>GSM4064976</th>\n",
       "      <th>GSM4064977</th>\n",
       "      <th>GSM4064978</th>\n",
       "      <th>GSM4064979</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM4065020</th>\n",
       "      <th>GSM4065021</th>\n",
       "      <th>GSM4065022</th>\n",
       "      <th>GSM4065023</th>\n",
       "      <th>GSM4065024</th>\n",
       "      <th>GSM4065025</th>\n",
       "      <th>GSM4065026</th>\n",
       "      <th>GSM4065027</th>\n",
       "      <th>GSM4065028</th>\n",
       "      <th>GSM4065029</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ILMN_1343291</th>\n",
       "      <td>9204.301266</td>\n",
       "      <td>5075.171525</td>\n",
       "      <td>10051.290505</td>\n",
       "      <td>12123.697693</td>\n",
       "      <td>10574.956805</td>\n",
       "      <td>12959.006463</td>\n",
       "      <td>4846.616019</td>\n",
       "      <td>15194.394130</td>\n",
       "      <td>4235.706279</td>\n",
       "      <td>6740.578348</td>\n",
       "      <td>...</td>\n",
       "      <td>15830.340787</td>\n",
       "      <td>15243.439606</td>\n",
       "      <td>14769.689168</td>\n",
       "      <td>13909.008710</td>\n",
       "      <td>10192.845233</td>\n",
       "      <td>14354.926566</td>\n",
       "      <td>16041.994931</td>\n",
       "      <td>13428.226366</td>\n",
       "      <td>1946.162525</td>\n",
       "      <td>11239.665798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651209</th>\n",
       "      <td>4.004890</td>\n",
       "      <td>5.342496</td>\n",
       "      <td>6.945065</td>\n",
       "      <td>8.094816</td>\n",
       "      <td>7.528906</td>\n",
       "      <td>4.823480</td>\n",
       "      <td>4.631520</td>\n",
       "      <td>9.696207</td>\n",
       "      <td>9.905800</td>\n",
       "      <td>8.640280</td>\n",
       "      <td>...</td>\n",
       "      <td>11.449560</td>\n",
       "      <td>24.699160</td>\n",
       "      <td>19.547560</td>\n",
       "      <td>40.412700</td>\n",
       "      <td>14.336090</td>\n",
       "      <td>22.246680</td>\n",
       "      <td>55.300770</td>\n",
       "      <td>112.238000</td>\n",
       "      <td>117.986800</td>\n",
       "      <td>16.114540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651228</th>\n",
       "      <td>153.984000</td>\n",
       "      <td>3792.218000</td>\n",
       "      <td>710.980200</td>\n",
       "      <td>5711.639000</td>\n",
       "      <td>114.615400</td>\n",
       "      <td>934.037900</td>\n",
       "      <td>20490.110000</td>\n",
       "      <td>8772.020000</td>\n",
       "      <td>934.037900</td>\n",
       "      <td>620.736600</td>\n",
       "      <td>...</td>\n",
       "      <td>99.317010</td>\n",
       "      <td>4045.860000</td>\n",
       "      <td>288.874000</td>\n",
       "      <td>490.598300</td>\n",
       "      <td>1090.540000</td>\n",
       "      <td>531.987500</td>\n",
       "      <td>1802.528000</td>\n",
       "      <td>956.679300</td>\n",
       "      <td>11.410130</td>\n",
       "      <td>637.642700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651229</th>\n",
       "      <td>4758.649323</td>\n",
       "      <td>4885.010970</td>\n",
       "      <td>6980.379696</td>\n",
       "      <td>3278.166966</td>\n",
       "      <td>4653.547573</td>\n",
       "      <td>4973.441489</td>\n",
       "      <td>591.788287</td>\n",
       "      <td>41.117053</td>\n",
       "      <td>1813.969520</td>\n",
       "      <td>4783.730491</td>\n",
       "      <td>...</td>\n",
       "      <td>5714.604457</td>\n",
       "      <td>5482.568242</td>\n",
       "      <td>5060.292222</td>\n",
       "      <td>6118.734244</td>\n",
       "      <td>7730.961740</td>\n",
       "      <td>6244.711505</td>\n",
       "      <td>5705.128276</td>\n",
       "      <td>7137.665780</td>\n",
       "      <td>8704.834259</td>\n",
       "      <td>7614.901015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ILMN_1651235</th>\n",
       "      <td>204.614435</td>\n",
       "      <td>51.636499</td>\n",
       "      <td>7.295319</td>\n",
       "      <td>6.614813</td>\n",
       "      <td>310.366424</td>\n",
       "      <td>1395.234969</td>\n",
       "      <td>236.749618</td>\n",
       "      <td>128.245911</td>\n",
       "      <td>5.101588</td>\n",
       "      <td>524.190165</td>\n",
       "      <td>...</td>\n",
       "      <td>1187.466338</td>\n",
       "      <td>714.294061</td>\n",
       "      <td>1679.476338</td>\n",
       "      <td>1384.370474</td>\n",
       "      <td>1021.047075</td>\n",
       "      <td>1435.140896</td>\n",
       "      <td>591.893404</td>\n",
       "      <td>2622.778277</td>\n",
       "      <td>376.727496</td>\n",
       "      <td>1150.683846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GSM4064970   GSM4064971    GSM4064972    GSM4064973  \\\n",
       "ID                                                                   \n",
       "ILMN_1343291  9204.301266  5075.171525  10051.290505  12123.697693   \n",
       "ILMN_1651209     4.004890     5.342496      6.945065      8.094816   \n",
       "ILMN_1651228   153.984000  3792.218000    710.980200   5711.639000   \n",
       "ILMN_1651229  4758.649323  4885.010970   6980.379696   3278.166966   \n",
       "ILMN_1651235   204.614435    51.636499      7.295319      6.614813   \n",
       "\n",
       "                GSM4064974    GSM4064975    GSM4064976    GSM4064977  \\\n",
       "ID                                                                     \n",
       "ILMN_1343291  10574.956805  12959.006463   4846.616019  15194.394130   \n",
       "ILMN_1651209      7.528906      4.823480      4.631520      9.696207   \n",
       "ILMN_1651228    114.615400    934.037900  20490.110000   8772.020000   \n",
       "ILMN_1651229   4653.547573   4973.441489    591.788287     41.117053   \n",
       "ILMN_1651235    310.366424   1395.234969    236.749618    128.245911   \n",
       "\n",
       "               GSM4064978   GSM4064979  ...    GSM4065020    GSM4065021  \\\n",
       "ID                                      ...                               \n",
       "ILMN_1343291  4235.706279  6740.578348  ...  15830.340787  15243.439606   \n",
       "ILMN_1651209     9.905800     8.640280  ...     11.449560     24.699160   \n",
       "ILMN_1651228   934.037900   620.736600  ...     99.317010   4045.860000   \n",
       "ILMN_1651229  1813.969520  4783.730491  ...   5714.604457   5482.568242   \n",
       "ILMN_1651235     5.101588   524.190165  ...   1187.466338    714.294061   \n",
       "\n",
       "                GSM4065022    GSM4065023    GSM4065024    GSM4065025  \\\n",
       "ID                                                                     \n",
       "ILMN_1343291  14769.689168  13909.008710  10192.845233  14354.926566   \n",
       "ILMN_1651209     19.547560     40.412700     14.336090     22.246680   \n",
       "ILMN_1651228    288.874000    490.598300   1090.540000    531.987500   \n",
       "ILMN_1651229   5060.292222   6118.734244   7730.961740   6244.711505   \n",
       "ILMN_1651235   1679.476338   1384.370474   1021.047075   1435.140896   \n",
       "\n",
       "                GSM4065026    GSM4065027   GSM4065028    GSM4065029  \n",
       "ID                                                                   \n",
       "ILMN_1343291  16041.994931  13428.226366  1946.162525  11239.665798  \n",
       "ILMN_1651209     55.300770    112.238000   117.986800     16.114540  \n",
       "ILMN_1651228   1802.528000    956.679300    11.410130    637.642700  \n",
       "ILMN_1651229   5705.128276   7137.665780  8704.834259   7614.901015  \n",
       "ILMN_1651235    591.893404   2622.778277   376.727496   1150.683846  \n",
       "\n",
       "[5 rows x 60 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:06.250833943Z",
     "start_time": "2024-01-02T01:03:06.246758837Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ILMN_1343291',\n",
       " 'ILMN_1651209',\n",
       " 'ILMN_1651228',\n",
       " 'ILMN_1651229',\n",
       " 'ILMN_1651235',\n",
       " 'ILMN_1651236',\n",
       " 'ILMN_1651237',\n",
       " 'ILMN_1651238',\n",
       " 'ILMN_1651254',\n",
       " 'ILMN_1651260',\n",
       " 'ILMN_1651262',\n",
       " 'ILMN_1651268',\n",
       " 'ILMN_1651278',\n",
       " 'ILMN_1651282',\n",
       " 'ILMN_1651285',\n",
       " 'ILMN_1651286',\n",
       " 'ILMN_1651292',\n",
       " 'ILMN_1651303',\n",
       " 'ILMN_1651309',\n",
       " 'ILMN_1651315']"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:07.586216481Z",
     "start_time": "2024-01-02T01:03:07.574527035Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['ILMN_1343291', 'ILMN_1651209', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651235', 'ILMN_1651236', 'ILMN_1651237', 'ILMN_1651238', 'ILMN_1651254', 'ILMN_1651260', 'ILMN_1651262', 'ILMN_1651268', 'ILMN_1651278', 'ILMN_1651282', 'ILMN_1651285', 'ILMN_1651286', 'ILMN_1651292', 'ILMN_1651303', 'ILMN_1651309', 'ILMN_1651315']\\n\""
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:08.263712538Z",
     "start_time": "2024-01-02T01:03:08.260656314Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:33.822550310Z",
     "start_time": "2024-01-02T01:03:08.648958912Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['ILMN_3166687', 'ILMN_3165566', 'ILMN_3164811', 'ILMN_3165363', 'ILMN_3166511'], 'Transcript': ['ILMN_333737', 'ILMN_333646', 'ILMN_333584', 'ILMN_333628', 'ILMN_333719'], 'Species': ['ILMN Controls', 'ILMN Controls', 'ILMN Controls', 'ILMN Controls', 'ILMN Controls'], 'Source': ['ILMN_Controls', 'ILMN_Controls', 'ILMN_Controls', 'ILMN_Controls', 'ILMN_Controls'], 'Search_Key': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'ILMN_Gene': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'Source_Reference_ID': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': ['DQ516750', 'DQ883654', 'DQ668364', 'DQ516785', 'DQ854995'], 'Symbol': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'Protein_Product': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5270161.0, 4260594.0, 7610424.0, 5260356.0, 2030196.0], 'Probe_Type': ['S', 'S', 'S', 'S', 'S'], 'Probe_Start': [12.0, 224.0, 868.0, 873.0, 130.0], 'SEQUENCE': ['CCCATGTGTCCAATTCTGAATATCTTTCCAGCTAAGTGCTTCTGCCCACC', 'GGATTAACTGCTGTGGTGTGTCATACTCGGCTACCTCCTGGTTTGGCGTC', 'GACCACGCCTTGTAATCGTATGACACGCGCTTGACACGACTGAATCCAGC', 'CTGCAATGCCATTAACAACCTTAGCACGGTATTTCCAGTAGCTGGTGAGC', 'CGTGCAGACAGGGATCGTAAGGCGATCCAGCCGGTATACCTTAGTCACAT'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': ['Methanocaldococcus jannaschii spike-in control MJ-500-33 genomic sequence', 'Synthetic construct clone NISTag13 external RNA control sequence', 'Synthetic construct clone TagJ microarray control', 'Methanocaldococcus jannaschii spike-in control MJ-1000-68 genomic sequence', 'Synthetic construct clone AG006.1100 external RNA control sequence'], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': ['DQ516750', 'DQ883654', 'DQ668364', 'DQ516785', 'DQ854995']}\n"
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:33.822912446Z",
     "start_time": "2024-01-02T01:03:33.804491826Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n",
      "    probe_name_key = 'key_name1'\n",
      "    gene_name_key = 'key_name2'\n",
      "\n",
      "    Gene annotation dictionary:\n",
      "    {'ID': ['ILMN_3166687', 'ILMN_3165566', 'ILMN_3164811', 'ILMN_3165363', 'ILMN_3166511'], 'Transcript': ['ILMN_333737', 'ILMN_333646', 'ILMN_333584', 'ILMN_333628', 'ILMN_333719'], 'Species': ['ILMN Controls', 'ILMN Controls', 'ILMN Controls', 'ILMN Controls', 'ILMN Controls'], 'Source': ['ILMN_Controls', 'ILMN_Controls', 'ILMN_Controls', 'ILMN_Controls', 'ILMN_Controls'], 'Search_Key': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'ILMN_Gene': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'Source_Reference_ID': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': ['DQ516750', 'DQ883654', 'DQ668364', 'DQ516785', 'DQ854995'], 'Symbol': ['ERCC-00162', 'ERCC-00071', 'ERCC-00009', 'ERCC-00053', 'ERCC-00144'], 'Protein_Product': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5270161.0, 4260594.0, 7610424.0, 5260356.0, 2030196.0], 'Probe_Type': ['S', 'S', 'S', 'S', 'S'], 'Probe_Start': [12.0, 224.0, 868.0, 873.0, 130.0], 'SEQUENCE': ['CCCATGTGTCCAATTCTGAATATCTTTCCAGCTAAGTGCTTCTGCCCACC', 'GGATTAACTGCTGTGGTGTGTCATACTCGGCTACCTCCTGGTTTGGCGTC', 'GACCACGCCTTGTAATCGTATGACACGCGCTTGACACGACTGAATCCAGC', 'CTGCAATGCCATTAACAACCTTAGCACGGTATTTCCAGTAGCTGGTGAGC', 'CGTGCAGACAGGGATCGTAAGGCGATCCAGCCGGTATACCTTAGTCACAT'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': ['Methanocaldococcus jannaschii spike-in control MJ-500-33 genomic sequence', 'Synthetic construct clone NISTag13 external RNA control sequence', 'Synthetic construct clone TagJ microarray control', 'Methanocaldococcus jannaschii spike-in control MJ-1000-68 genomic sequence', 'Synthetic construct clone AG006.1100 external RNA control sequence'], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': ['DQ516750', 'DQ883654', 'DQ668364', 'DQ516785', 'DQ854995']}\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n",
    "    probe_name_key = 'key_name1'\n",
    "    gene_name_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:35.113359369Z",
     "start_time": "2024-01-02T01:03:33.804666371Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    probe_id_key = 'ID'\n",
    "    gene_symb_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, probe_id_key, gene_symb_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:35.156584675Z",
     "start_time": "2024-01-02T01:03:35.114652593Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "12 input query terms found dup hits:\t[('ABCC13', 2), ('ABCC6P1', 2), ('ABCC6P2', 3), ('ADAM6', 3), ('AGAP11', 2), ('ALOX12P2', 2), ('ANXA\n",
      "91 input query terms found no hit:\t['7A5', 'A26C3', 'A2BP1', 'A2LD1', 'AADACL1', 'AARS', 'ABCA11', 'ABHD7', 'ABHD9', 'ABP1', 'ACCN1', '\n",
      "4 input query terms found dup hits:\t[('ATP6AP1L', 2), ('ATXN8OS', 2), ('BAGE2', 2), ('BRD7P3', 2)]\n",
      "395 input query terms found no hit:\t['ARS2', 'ARSE', 'ARVP6125', 'ASAM', 'ASAP1IT1', 'ASFMR1', 'ASNA1', 'ATHL1', 'ATP5A1', 'ATP5B', 'ATP\n",
      "2 input query terms found dup hits:\t[('C2orf83', 2), ('C5orf60', 2)]\n",
      "638 input query terms found no hit:\t['C17orf72', 'C17orf73', 'C17orf74', 'C17orf76', 'C17orf77', 'C17orf79', 'C17orf80', 'C17orf81', 'C1\n",
      "4 input query terms found dup hits:\t[('CCDC144NL', 2), ('CCT6P1', 2), ('CDR1', 2), ('CLEC4GP1', 2)]\n",
      "113 input query terms found no hit:\t['CCBL1', 'CCBL2', 'CCBP2', 'CCDC101', 'CCDC104', 'CCDC105', 'CCDC108', 'CCDC109A', 'CCDC109B', 'CCD\n",
      "8 input query terms found dup hits:\t[('COPG2IT1', 2), ('CXADRP2', 3), ('CXADRP3', 2), ('CYP4Z2P', 2), ('DEFB122', 2), ('DHRS4L1', 3), ('\n",
      "120 input query terms found no hit:\t['COPG', 'COX11P', 'COX4NB', 'CP110', 'CPSF3L', 'CRAMP1L', 'CREBL1', 'CRIPAK', 'CRKRS', 'CROCCL2', '\n",
      "4 input query terms found dup hits:\t[('DNM1P35', 2), ('DPRXP4', 3), ('DPY19L2P1', 2), ('DUXAP3', 2)]\n",
      "248 input query terms found no hit:\t['DNHD2', 'DNHL1', 'DOM3Z', 'DOPEY1', 'DOPEY2', 'DPCR1', 'DRD1IP', 'DSCR1L1', 'DSCR3', 'DSCR5', 'DSC\n",
      "8 input query terms found dup hits:\t[('FAM41AY2', 2), ('FAM74A3', 2), ('FCGR2C', 2), ('FER1L4', 2), ('FMO9P', 2), ('FOLH1B', 2), ('GABPA\n",
      "322 input query terms found no hit:\t['FAM189B', 'FAM18A', 'FAM18B', 'FAM18B2', 'FAM190B', 'FAM194A', 'FAM195A', 'FAM195B', 'FAM19A1', 'F\n",
      "11 input query terms found dup hits:\t[('GGT3P', 2), ('GGT8P', 2), ('GLRA4', 2), ('GTF2H2B', 2), ('HBBP1', 2), ('HBD', 2), ('HCG4', 9), ('\n",
      "188 input query terms found no hit:\t['GCN1L1', 'GDEP', 'GEFT', 'GIF', 'GIYD1', 'GIYD2', 'GLT25D1', 'GLT25D2', 'GLTPD1', 'GLTSCR1', 'GLTS\n",
      "7 input query terms found dup hits:\t[('HNRNPA3P1', 2), ('HSD17B7P2', 2), ('HSP90AB2P', 2), ('HSP90AB4P', 2), ('HSP90B3P', 2), ('HSPA7', \n",
      "130 input query terms found no hit:\t['HMP19', 'HN1', 'HN1L', 'HNRPA1L-2', 'HNRPA2B1', 'HNRPC', 'HNRPDL', 'HNRPH1', 'HNRPH3', 'HNRPK', 'H\n",
      "3 input query terms found dup hits:\t[('KIR3DP1', 10), ('KIR3DX1', 5), ('KLKP1', 2)]\n",
      "385 input query terms found no hit:\t['KIAA0649', 'KIAA0652', 'KIAA0664', 'KIAA0672', 'KIAA0773', 'KIAA0776', 'KIAA0802', 'KIAA0831', 'KI\n",
      "7 input query terms found dup hits:\t[('LPAL2', 2), ('LRP5L', 2), ('LY6G6E', 8), ('LYPLA2P1', 6), ('MALAT1', 2), ('MEG8', 2), ('MEIS3P1',\n",
      "318 input query terms found no hit:\t['LOC441120', 'LOC441150', 'LOC441177', 'LOC441178', 'LOC441204', 'LOC441208', 'LOC441233', 'LOC4412\n",
      "13 input query terms found dup hits:\t[('MIR320A', 2), ('MIR320B1', 2), ('MIR320B2', 2), ('MIR320C1', 2), ('MMD2', 2), ('MMP23A', 2), ('MO\n",
      "85 input query terms found no hit:\t['MIR1245', 'MIR1269', 'MIR1274A', 'MIR1274B', 'MIR1295', 'MIR1308', 'MIR147', 'MIR1826', 'MIR190', \n",
      "11 input query terms found dup hits:\t[('NAPSB', 2), ('NAT8B', 3), ('NBPF22P', 2), ('NCF1B', 2), ('NME2P1', 2), ('NUDT9P1', 2), ('NXF4', 2\n",
      "110 input query terms found no hit:\t['MYOHD1', 'MYST1', 'MYST2', 'MYST3', 'MYST4', 'Magmas', 'March 1', 'March 10', 'March 11', 'March 2\n",
      "8 input query terms found dup hits:\t[('OR6W1P', 3), ('OR7E156P', 2), ('OR7E37P', 2), ('OR7E91P', 2), ('P2RX6P', 2), ('PA2G4P4', 2), ('PD\n",
      "73 input query terms found no hit:\t['OR2W5', 'OR3A4', 'OR52A4', 'OR5AY1', 'OR5R1', 'OR8G2', 'ORAOV1', 'ORC1L', 'ORC2L', 'ORC3L', 'ORC4L\n",
      "11 input query terms found dup hits:\t[('PMCHL1', 2), ('PMS2CL', 2), ('POLR2J3', 2), ('POM121L10P', 2), ('POM121L1P', 2), ('POM121L4P', 2)\n",
      "90 input query terms found no hit:\t['PLA2G16', 'PLAC1L', 'PLAC2', 'PLDN', 'PLEC1', 'PLEKHA9', 'PLIN', 'PLK5P', 'PLUNC', 'PLXNA4B', 'PMS\n",
      "17 input query terms found dup hits:\t[('RAET1K', 2), ('RAP1BL', 2), ('RBMY1A3P', 3), ('RBMY2EP', 2), ('RBMY2FP', 2), ('RBMY3AP', 3), ('RE\n",
      "81 input query terms found no hit:\t['RAB7L1', 'RABL4', 'RABL5', 'RAD51L1', 'RAD51L3', 'RAG1AP1', 'RAGE', 'RARRES3', 'RARS', 'RASA4P', '\n",
      "9 input query terms found dup hits:\t[('SCARNA17', 2), ('SCARNA4', 2), ('SDC4P', 2), ('SDHAP2', 9), ('SDHAP3', 2), ('SFTA3', 2), ('SIGLEC\n",
      "82 input query terms found no hit:\t['SCGBL', 'SCXA', 'SCXB', 'SCYE1', 'SCYL1BP1', 'SDCCAG1', 'SDCCAG10', 'SDCCAG3', 'SDPR', 'SEC1', 'SE\n",
      "64 input query terms found dup hits:\t[('SNHG5', 2), ('SNHG6', 2), ('SNHG8', 2), ('SNHG9', 3), ('SNORA16A', 2), ('SNORA21', 2), ('SNORA40'\n",
      "72 input query terms found no hit:\t['SNHG3-RCC1', 'SNIP', 'SNORA17', 'SNORA3', 'SNORA34', 'SNORA39', 'SNORA42', 'SNORA43', 'SNORA45', '\n",
      "8 input query terms found dup hits:\t[('TCP10L2', 2), ('TEC', 2), ('TERC', 2), ('TFAMP1', 2), ('TNXA', 5), ('TOP1P1', 2), ('TP53TG1', 2),\n",
      "104 input query terms found no hit:\t['TARP', 'TARS', 'TARSL2', 'TAS2R49', 'TAZ', 'TBC1D29', 'TCAM1', 'TCBA1', 'TCEB1', 'TCEB2', 'TCEB3',\n",
      "12 input query terms found dup hits:\t[('TRPC2', 2), ('TTTY11', 2), ('TTTY17C', 2), ('TTTY18', 2), ('TTTY2', 2), ('TYRO3P', 2), ('UBE2MP1'\n",
      "124 input query terms found no hit:\t['TRIM78P', 'TROVE2', 'TRY1', 'TRY6', 'TRYX3', 'TSGA14', 'TSP50', 'TSPYL3', 'TSSC1', 'TSTA3', 'TTC15\n",
      "3 input query terms found dup hits:\t[('XKRY', 2), ('ZNF658B', 2), ('ZNF781', 2)]\n",
      "62 input query terms found no hit:\t['XKRY2', 'XLKD1', 'XRCC6BP1', 'YARS', 'YSK4', 'ZADH2', 'ZAK', 'ZCCHC11', 'ZCCHC16', 'ZCCHC5', 'ZCCH\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:35.158174021Z",
     "start_time": "2024-01-02T01:03:35.156777134Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:35.158378853Z",
     "start_time": "2024-01-02T01:03:35.156895194Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 60 samples.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Brugada Syndrome</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM4064970</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62.355348</td>\n",
       "      <td>14.440420</td>\n",
       "      <td>2460.196694</td>\n",
       "      <td>4.686300</td>\n",
       "      <td>6.105481</td>\n",
       "      <td>587.097412</td>\n",
       "      <td>16.382563</td>\n",
       "      <td>...</td>\n",
       "      <td>475.836640</td>\n",
       "      <td>10.585559</td>\n",
       "      <td>1858.563000</td>\n",
       "      <td>309.923000</td>\n",
       "      <td>238.824645</td>\n",
       "      <td>16.280420</td>\n",
       "      <td>27.983100</td>\n",
       "      <td>925.877535</td>\n",
       "      <td>457.178400</td>\n",
       "      <td>1711.636700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064971</th>\n",
       "      <td>0.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.278570</td>\n",
       "      <td>7.569081</td>\n",
       "      <td>7293.759251</td>\n",
       "      <td>5.559912</td>\n",
       "      <td>4.979804</td>\n",
       "      <td>327.278673</td>\n",
       "      <td>6.757537</td>\n",
       "      <td>...</td>\n",
       "      <td>114.519993</td>\n",
       "      <td>10.859937</td>\n",
       "      <td>452.635850</td>\n",
       "      <td>432.893700</td>\n",
       "      <td>342.556525</td>\n",
       "      <td>6.001372</td>\n",
       "      <td>328.690900</td>\n",
       "      <td>323.831265</td>\n",
       "      <td>1071.818000</td>\n",
       "      <td>1576.779250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064972</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54.362789</td>\n",
       "      <td>5.384082</td>\n",
       "      <td>1536.235125</td>\n",
       "      <td>195.513335</td>\n",
       "      <td>65.989054</td>\n",
       "      <td>2739.802065</td>\n",
       "      <td>109.323537</td>\n",
       "      <td>...</td>\n",
       "      <td>907.482433</td>\n",
       "      <td>25.361408</td>\n",
       "      <td>2418.226000</td>\n",
       "      <td>6.477754</td>\n",
       "      <td>6.872454</td>\n",
       "      <td>10.014170</td>\n",
       "      <td>182.508200</td>\n",
       "      <td>1914.635812</td>\n",
       "      <td>6.965921</td>\n",
       "      <td>1755.359300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064973</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.764988</td>\n",
       "      <td>8.260571</td>\n",
       "      <td>4517.352275</td>\n",
       "      <td>5.725376</td>\n",
       "      <td>6.391028</td>\n",
       "      <td>806.762413</td>\n",
       "      <td>45.303023</td>\n",
       "      <td>...</td>\n",
       "      <td>492.995270</td>\n",
       "      <td>13.643631</td>\n",
       "      <td>568.066500</td>\n",
       "      <td>3557.904000</td>\n",
       "      <td>933.398150</td>\n",
       "      <td>7.851095</td>\n",
       "      <td>308.473100</td>\n",
       "      <td>1901.632360</td>\n",
       "      <td>165.038800</td>\n",
       "      <td>1326.673000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064974</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.992323</td>\n",
       "      <td>18.493197</td>\n",
       "      <td>4215.080680</td>\n",
       "      <td>17.109983</td>\n",
       "      <td>7.504535</td>\n",
       "      <td>923.792662</td>\n",
       "      <td>9.095975</td>\n",
       "      <td>...</td>\n",
       "      <td>449.435800</td>\n",
       "      <td>13.533547</td>\n",
       "      <td>1767.799450</td>\n",
       "      <td>235.760200</td>\n",
       "      <td>122.383625</td>\n",
       "      <td>45.240360</td>\n",
       "      <td>30.014700</td>\n",
       "      <td>901.722045</td>\n",
       "      <td>352.542300</td>\n",
       "      <td>1327.331950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064975</th>\n",
       "      <td>0.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>31.366236</td>\n",
       "      <td>5.428562</td>\n",
       "      <td>1180.786270</td>\n",
       "      <td>5.852868</td>\n",
       "      <td>22.394404</td>\n",
       "      <td>958.550157</td>\n",
       "      <td>73.304113</td>\n",
       "      <td>...</td>\n",
       "      <td>901.280519</td>\n",
       "      <td>16.117974</td>\n",
       "      <td>2119.380000</td>\n",
       "      <td>762.076200</td>\n",
       "      <td>391.236405</td>\n",
       "      <td>9.205301</td>\n",
       "      <td>4.563783</td>\n",
       "      <td>2633.665940</td>\n",
       "      <td>5.055491</td>\n",
       "      <td>926.780490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064976</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>267.552914</td>\n",
       "      <td>8.712279</td>\n",
       "      <td>8764.367836</td>\n",
       "      <td>37.777003</td>\n",
       "      <td>438.316512</td>\n",
       "      <td>4.904852</td>\n",
       "      <td>6.736836</td>\n",
       "      <td>...</td>\n",
       "      <td>231.004713</td>\n",
       "      <td>11.046941</td>\n",
       "      <td>401.668940</td>\n",
       "      <td>7046.628000</td>\n",
       "      <td>244.577066</td>\n",
       "      <td>3.651659</td>\n",
       "      <td>2042.331000</td>\n",
       "      <td>1020.103037</td>\n",
       "      <td>162.418400</td>\n",
       "      <td>8.467687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064977</th>\n",
       "      <td>0.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1879.845723</td>\n",
       "      <td>287.377749</td>\n",
       "      <td>7.111364</td>\n",
       "      <td>5.390598</td>\n",
       "      <td>245.963630</td>\n",
       "      <td>80.346495</td>\n",
       "      <td>1189.334982</td>\n",
       "      <td>...</td>\n",
       "      <td>320.578720</td>\n",
       "      <td>10.155646</td>\n",
       "      <td>733.606260</td>\n",
       "      <td>7.825502</td>\n",
       "      <td>93.016329</td>\n",
       "      <td>219.472800</td>\n",
       "      <td>7.363958</td>\n",
       "      <td>8311.804700</td>\n",
       "      <td>4.340990</td>\n",
       "      <td>169.654780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064978</th>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>112.182100</td>\n",
       "      <td>7.417203</td>\n",
       "      <td>3260.866611</td>\n",
       "      <td>7.106071</td>\n",
       "      <td>10.197472</td>\n",
       "      <td>268.349773</td>\n",
       "      <td>17.072828</td>\n",
       "      <td>...</td>\n",
       "      <td>369.459033</td>\n",
       "      <td>8.997515</td>\n",
       "      <td>181.475945</td>\n",
       "      <td>11267.170000</td>\n",
       "      <td>1351.171000</td>\n",
       "      <td>7.917643</td>\n",
       "      <td>126.146200</td>\n",
       "      <td>3719.357300</td>\n",
       "      <td>241.948900</td>\n",
       "      <td>887.739150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064979</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>78.592842</td>\n",
       "      <td>30.706918</td>\n",
       "      <td>1559.723056</td>\n",
       "      <td>9.450259</td>\n",
       "      <td>10.040594</td>\n",
       "      <td>611.063243</td>\n",
       "      <td>74.828051</td>\n",
       "      <td>...</td>\n",
       "      <td>644.490500</td>\n",
       "      <td>11.755870</td>\n",
       "      <td>2287.628500</td>\n",
       "      <td>26.475930</td>\n",
       "      <td>189.163285</td>\n",
       "      <td>12.040690</td>\n",
       "      <td>18.647120</td>\n",
       "      <td>1397.042660</td>\n",
       "      <td>4.680267</td>\n",
       "      <td>1564.791560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064980</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.500891</td>\n",
       "      <td>8.117416</td>\n",
       "      <td>6556.304666</td>\n",
       "      <td>3.985734</td>\n",
       "      <td>7.040947</td>\n",
       "      <td>226.164405</td>\n",
       "      <td>11.574808</td>\n",
       "      <td>...</td>\n",
       "      <td>143.642440</td>\n",
       "      <td>10.698736</td>\n",
       "      <td>213.605250</td>\n",
       "      <td>606.987800</td>\n",
       "      <td>294.037245</td>\n",
       "      <td>8.095774</td>\n",
       "      <td>164.823500</td>\n",
       "      <td>476.196355</td>\n",
       "      <td>971.513800</td>\n",
       "      <td>669.199000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064981</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.975764</td>\n",
       "      <td>7.315140</td>\n",
       "      <td>3293.237621</td>\n",
       "      <td>6.412876</td>\n",
       "      <td>5.016920</td>\n",
       "      <td>495.804636</td>\n",
       "      <td>8.940770</td>\n",
       "      <td>...</td>\n",
       "      <td>489.470387</td>\n",
       "      <td>60.020860</td>\n",
       "      <td>895.049800</td>\n",
       "      <td>723.229200</td>\n",
       "      <td>474.962190</td>\n",
       "      <td>15.774970</td>\n",
       "      <td>204.312500</td>\n",
       "      <td>622.720075</td>\n",
       "      <td>366.868200</td>\n",
       "      <td>1844.402700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064982</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.740504</td>\n",
       "      <td>9.078905</td>\n",
       "      <td>3562.690498</td>\n",
       "      <td>25.182167</td>\n",
       "      <td>9.602690</td>\n",
       "      <td>148.658069</td>\n",
       "      <td>12.069092</td>\n",
       "      <td>...</td>\n",
       "      <td>291.751847</td>\n",
       "      <td>28.721025</td>\n",
       "      <td>646.298115</td>\n",
       "      <td>5751.135000</td>\n",
       "      <td>1758.740500</td>\n",
       "      <td>25.433950</td>\n",
       "      <td>200.265300</td>\n",
       "      <td>537.587265</td>\n",
       "      <td>278.035600</td>\n",
       "      <td>375.949015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064983</th>\n",
       "      <td>0.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>236.178258</td>\n",
       "      <td>38.379484</td>\n",
       "      <td>5261.985908</td>\n",
       "      <td>4.361902</td>\n",
       "      <td>174.406763</td>\n",
       "      <td>2192.210826</td>\n",
       "      <td>231.662768</td>\n",
       "      <td>...</td>\n",
       "      <td>42.409687</td>\n",
       "      <td>23.656955</td>\n",
       "      <td>2197.597702</td>\n",
       "      <td>360.965800</td>\n",
       "      <td>16.988810</td>\n",
       "      <td>801.090400</td>\n",
       "      <td>63.027030</td>\n",
       "      <td>2246.607310</td>\n",
       "      <td>5.163648</td>\n",
       "      <td>2987.136878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064984</th>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.733671</td>\n",
       "      <td>15.222632</td>\n",
       "      <td>3034.285917</td>\n",
       "      <td>8.050865</td>\n",
       "      <td>8.015696</td>\n",
       "      <td>820.683189</td>\n",
       "      <td>14.982735</td>\n",
       "      <td>...</td>\n",
       "      <td>828.796156</td>\n",
       "      <td>128.576423</td>\n",
       "      <td>1711.594000</td>\n",
       "      <td>13.635600</td>\n",
       "      <td>64.651605</td>\n",
       "      <td>5.867572</td>\n",
       "      <td>47.443430</td>\n",
       "      <td>1135.455990</td>\n",
       "      <td>6.997004</td>\n",
       "      <td>729.671550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064985</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.728377</td>\n",
       "      <td>8.154470</td>\n",
       "      <td>8701.918648</td>\n",
       "      <td>4.693914</td>\n",
       "      <td>4.645250</td>\n",
       "      <td>575.679296</td>\n",
       "      <td>9.831937</td>\n",
       "      <td>...</td>\n",
       "      <td>536.601530</td>\n",
       "      <td>111.695540</td>\n",
       "      <td>555.470200</td>\n",
       "      <td>937.922200</td>\n",
       "      <td>469.344995</td>\n",
       "      <td>8.042021</td>\n",
       "      <td>187.039300</td>\n",
       "      <td>561.487255</td>\n",
       "      <td>690.213200</td>\n",
       "      <td>1559.428500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064986</th>\n",
       "      <td>0.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.679665</td>\n",
       "      <td>7.576425</td>\n",
       "      <td>5548.304952</td>\n",
       "      <td>9.535022</td>\n",
       "      <td>4.319101</td>\n",
       "      <td>757.620403</td>\n",
       "      <td>11.472898</td>\n",
       "      <td>...</td>\n",
       "      <td>512.308813</td>\n",
       "      <td>18.318939</td>\n",
       "      <td>1472.623500</td>\n",
       "      <td>820.456000</td>\n",
       "      <td>583.374900</td>\n",
       "      <td>4.111516</td>\n",
       "      <td>158.076200</td>\n",
       "      <td>704.123905</td>\n",
       "      <td>254.941600</td>\n",
       "      <td>1842.112400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064987</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>47.697543</td>\n",
       "      <td>151.769953</td>\n",
       "      <td>4480.836105</td>\n",
       "      <td>4.439174</td>\n",
       "      <td>6.913088</td>\n",
       "      <td>857.596072</td>\n",
       "      <td>51.583578</td>\n",
       "      <td>...</td>\n",
       "      <td>529.796164</td>\n",
       "      <td>27.621451</td>\n",
       "      <td>778.790050</td>\n",
       "      <td>1150.408000</td>\n",
       "      <td>293.628510</td>\n",
       "      <td>149.278600</td>\n",
       "      <td>8.218412</td>\n",
       "      <td>1373.463340</td>\n",
       "      <td>462.977200</td>\n",
       "      <td>1209.270365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064988</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>75.198431</td>\n",
       "      <td>6.857917</td>\n",
       "      <td>2615.503687</td>\n",
       "      <td>7.667636</td>\n",
       "      <td>7.647184</td>\n",
       "      <td>180.074405</td>\n",
       "      <td>8.011405</td>\n",
       "      <td>...</td>\n",
       "      <td>170.617420</td>\n",
       "      <td>80.205718</td>\n",
       "      <td>19.971890</td>\n",
       "      <td>12410.470000</td>\n",
       "      <td>1020.046650</td>\n",
       "      <td>8.103821</td>\n",
       "      <td>457.943000</td>\n",
       "      <td>1123.434000</td>\n",
       "      <td>262.584300</td>\n",
       "      <td>359.373570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064989</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>586.805849</td>\n",
       "      <td>111.214584</td>\n",
       "      <td>1404.553002</td>\n",
       "      <td>63.813658</td>\n",
       "      <td>5.573835</td>\n",
       "      <td>6.926686</td>\n",
       "      <td>8.391545</td>\n",
       "      <td>...</td>\n",
       "      <td>231.132710</td>\n",
       "      <td>35.340658</td>\n",
       "      <td>5.441949</td>\n",
       "      <td>11.570360</td>\n",
       "      <td>1434.869861</td>\n",
       "      <td>10.258190</td>\n",
       "      <td>1059.371000</td>\n",
       "      <td>5155.736470</td>\n",
       "      <td>4.640261</td>\n",
       "      <td>6.326132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064990</th>\n",
       "      <td>0.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>234.311653</td>\n",
       "      <td>5948.300181</td>\n",
       "      <td>14769.609246</td>\n",
       "      <td>8.206099</td>\n",
       "      <td>5.786520</td>\n",
       "      <td>9.875059</td>\n",
       "      <td>16.007956</td>\n",
       "      <td>...</td>\n",
       "      <td>186.685233</td>\n",
       "      <td>9.615675</td>\n",
       "      <td>352.192000</td>\n",
       "      <td>1318.780000</td>\n",
       "      <td>2260.991300</td>\n",
       "      <td>399.729500</td>\n",
       "      <td>183.506900</td>\n",
       "      <td>142.258435</td>\n",
       "      <td>1559.518000</td>\n",
       "      <td>1488.380950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064991</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.966833</td>\n",
       "      <td>1459.044789</td>\n",
       "      <td>9684.632839</td>\n",
       "      <td>5.403382</td>\n",
       "      <td>5.623221</td>\n",
       "      <td>99.371524</td>\n",
       "      <td>23.985891</td>\n",
       "      <td>...</td>\n",
       "      <td>648.563097</td>\n",
       "      <td>28.267225</td>\n",
       "      <td>292.516800</td>\n",
       "      <td>856.175300</td>\n",
       "      <td>478.450515</td>\n",
       "      <td>67.469250</td>\n",
       "      <td>40.998740</td>\n",
       "      <td>492.073800</td>\n",
       "      <td>545.807800</td>\n",
       "      <td>2002.681900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064992</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>142.618631</td>\n",
       "      <td>2712.974607</td>\n",
       "      <td>9394.966901</td>\n",
       "      <td>5.294361</td>\n",
       "      <td>6.947625</td>\n",
       "      <td>21.549403</td>\n",
       "      <td>8.514557</td>\n",
       "      <td>...</td>\n",
       "      <td>943.467097</td>\n",
       "      <td>64.065413</td>\n",
       "      <td>1153.041450</td>\n",
       "      <td>1286.365000</td>\n",
       "      <td>634.685975</td>\n",
       "      <td>472.844100</td>\n",
       "      <td>148.258000</td>\n",
       "      <td>834.830370</td>\n",
       "      <td>384.734100</td>\n",
       "      <td>2447.874200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064993</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.504939</td>\n",
       "      <td>1482.934434</td>\n",
       "      <td>11401.858638</td>\n",
       "      <td>4.534517</td>\n",
       "      <td>6.299148</td>\n",
       "      <td>122.005959</td>\n",
       "      <td>9.861370</td>\n",
       "      <td>...</td>\n",
       "      <td>563.999787</td>\n",
       "      <td>58.405770</td>\n",
       "      <td>683.971150</td>\n",
       "      <td>1100.959000</td>\n",
       "      <td>374.124685</td>\n",
       "      <td>251.940300</td>\n",
       "      <td>90.164050</td>\n",
       "      <td>811.157355</td>\n",
       "      <td>318.022700</td>\n",
       "      <td>1691.558100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064994</th>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.936492</td>\n",
       "      <td>3833.436124</td>\n",
       "      <td>13289.794414</td>\n",
       "      <td>4.225422</td>\n",
       "      <td>5.420832</td>\n",
       "      <td>25.827550</td>\n",
       "      <td>6.355564</td>\n",
       "      <td>...</td>\n",
       "      <td>280.885937</td>\n",
       "      <td>9.353346</td>\n",
       "      <td>306.606800</td>\n",
       "      <td>974.010300</td>\n",
       "      <td>515.183950</td>\n",
       "      <td>79.411110</td>\n",
       "      <td>116.111700</td>\n",
       "      <td>281.953865</td>\n",
       "      <td>662.150700</td>\n",
       "      <td>1502.183100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064995</th>\n",
       "      <td>0.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>726.484206</td>\n",
       "      <td>12.763040</td>\n",
       "      <td>5514.580124</td>\n",
       "      <td>166.227748</td>\n",
       "      <td>9.098739</td>\n",
       "      <td>4.948842</td>\n",
       "      <td>561.678828</td>\n",
       "      <td>...</td>\n",
       "      <td>30.660071</td>\n",
       "      <td>36.665401</td>\n",
       "      <td>132.786752</td>\n",
       "      <td>73.986010</td>\n",
       "      <td>2135.142185</td>\n",
       "      <td>125.304800</td>\n",
       "      <td>943.686500</td>\n",
       "      <td>973.182250</td>\n",
       "      <td>10.223680</td>\n",
       "      <td>38.197745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064996</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2159.771081</td>\n",
       "      <td>15.451472</td>\n",
       "      <td>3746.966295</td>\n",
       "      <td>19.563223</td>\n",
       "      <td>52.677832</td>\n",
       "      <td>243.517783</td>\n",
       "      <td>27.023779</td>\n",
       "      <td>...</td>\n",
       "      <td>183.798180</td>\n",
       "      <td>119.650187</td>\n",
       "      <td>128.729310</td>\n",
       "      <td>6.084738</td>\n",
       "      <td>119.029350</td>\n",
       "      <td>432.274800</td>\n",
       "      <td>673.329300</td>\n",
       "      <td>5206.168050</td>\n",
       "      <td>179.628800</td>\n",
       "      <td>380.590205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064997</th>\n",
       "      <td>0.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>38.209843</td>\n",
       "      <td>636.232491</td>\n",
       "      <td>10756.733813</td>\n",
       "      <td>5.291773</td>\n",
       "      <td>7.199933</td>\n",
       "      <td>135.180297</td>\n",
       "      <td>6.741027</td>\n",
       "      <td>...</td>\n",
       "      <td>497.643930</td>\n",
       "      <td>19.221363</td>\n",
       "      <td>333.777700</td>\n",
       "      <td>3602.606000</td>\n",
       "      <td>1090.534000</td>\n",
       "      <td>46.158280</td>\n",
       "      <td>50.135550</td>\n",
       "      <td>519.852540</td>\n",
       "      <td>887.203800</td>\n",
       "      <td>1722.465000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064998</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>213.324275</td>\n",
       "      <td>1325.104073</td>\n",
       "      <td>12237.567394</td>\n",
       "      <td>5.526838</td>\n",
       "      <td>5.766445</td>\n",
       "      <td>25.970225</td>\n",
       "      <td>22.744779</td>\n",
       "      <td>...</td>\n",
       "      <td>840.886717</td>\n",
       "      <td>134.264273</td>\n",
       "      <td>419.692650</td>\n",
       "      <td>223.958500</td>\n",
       "      <td>487.167450</td>\n",
       "      <td>458.851300</td>\n",
       "      <td>43.845360</td>\n",
       "      <td>528.930775</td>\n",
       "      <td>6.546703</td>\n",
       "      <td>1856.440850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4064999</th>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.643003</td>\n",
       "      <td>854.772741</td>\n",
       "      <td>5.534308</td>\n",
       "      <td>15.926672</td>\n",
       "      <td>2955.620292</td>\n",
       "      <td>5.534308</td>\n",
       "      <td>222.287611</td>\n",
       "      <td>...</td>\n",
       "      <td>3575.796034</td>\n",
       "      <td>2550.317877</td>\n",
       "      <td>588.894001</td>\n",
       "      <td>172.472900</td>\n",
       "      <td>6.128139</td>\n",
       "      <td>5045.076000</td>\n",
       "      <td>264.535500</td>\n",
       "      <td>11.737111</td>\n",
       "      <td>5999.011000</td>\n",
       "      <td>149.815400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065000</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>115.095751</td>\n",
       "      <td>1013.827503</td>\n",
       "      <td>7876.024509</td>\n",
       "      <td>8.166187</td>\n",
       "      <td>17.061461</td>\n",
       "      <td>257.494382</td>\n",
       "      <td>80.656827</td>\n",
       "      <td>...</td>\n",
       "      <td>753.020108</td>\n",
       "      <td>15.955018</td>\n",
       "      <td>455.715100</td>\n",
       "      <td>2080.280000</td>\n",
       "      <td>1598.767650</td>\n",
       "      <td>907.107300</td>\n",
       "      <td>6.655668</td>\n",
       "      <td>1513.037035</td>\n",
       "      <td>492.228700</td>\n",
       "      <td>1258.421950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.668886</td>\n",
       "      <td>5410.263027</td>\n",
       "      <td>15126.746543</td>\n",
       "      <td>4.398537</td>\n",
       "      <td>7.060038</td>\n",
       "      <td>8.176693</td>\n",
       "      <td>10.599370</td>\n",
       "      <td>...</td>\n",
       "      <td>477.573507</td>\n",
       "      <td>175.482793</td>\n",
       "      <td>163.191750</td>\n",
       "      <td>365.437300</td>\n",
       "      <td>1327.685945</td>\n",
       "      <td>140.720900</td>\n",
       "      <td>134.320000</td>\n",
       "      <td>319.536445</td>\n",
       "      <td>795.299100</td>\n",
       "      <td>1165.559900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065002</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.398669</td>\n",
       "      <td>871.941258</td>\n",
       "      <td>8889.129204</td>\n",
       "      <td>8.516182</td>\n",
       "      <td>6.951950</td>\n",
       "      <td>834.090544</td>\n",
       "      <td>14.354971</td>\n",
       "      <td>...</td>\n",
       "      <td>1132.310457</td>\n",
       "      <td>219.915290</td>\n",
       "      <td>649.928000</td>\n",
       "      <td>2584.695000</td>\n",
       "      <td>514.736750</td>\n",
       "      <td>143.806900</td>\n",
       "      <td>61.196830</td>\n",
       "      <td>1258.490470</td>\n",
       "      <td>412.119300</td>\n",
       "      <td>1642.838750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065003</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.387492</td>\n",
       "      <td>3399.569849</td>\n",
       "      <td>14975.615481</td>\n",
       "      <td>5.865120</td>\n",
       "      <td>9.070901</td>\n",
       "      <td>16.649167</td>\n",
       "      <td>5.892301</td>\n",
       "      <td>...</td>\n",
       "      <td>403.493293</td>\n",
       "      <td>92.251820</td>\n",
       "      <td>234.376500</td>\n",
       "      <td>835.752100</td>\n",
       "      <td>493.296640</td>\n",
       "      <td>162.268600</td>\n",
       "      <td>74.689590</td>\n",
       "      <td>320.661088</td>\n",
       "      <td>497.860700</td>\n",
       "      <td>1677.421700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>784.652359</td>\n",
       "      <td>2326.449789</td>\n",
       "      <td>8997.131522</td>\n",
       "      <td>6.664390</td>\n",
       "      <td>9.163141</td>\n",
       "      <td>8.417904</td>\n",
       "      <td>10.878410</td>\n",
       "      <td>...</td>\n",
       "      <td>410.482363</td>\n",
       "      <td>14.841379</td>\n",
       "      <td>697.120800</td>\n",
       "      <td>3235.150000</td>\n",
       "      <td>2570.936850</td>\n",
       "      <td>5.946879</td>\n",
       "      <td>157.209600</td>\n",
       "      <td>1014.134508</td>\n",
       "      <td>4.868951</td>\n",
       "      <td>1663.368800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065005</th>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1600.291058</td>\n",
       "      <td>1896.662925</td>\n",
       "      <td>8895.637507</td>\n",
       "      <td>18.668081</td>\n",
       "      <td>6.935069</td>\n",
       "      <td>49.675526</td>\n",
       "      <td>352.678070</td>\n",
       "      <td>...</td>\n",
       "      <td>93.483760</td>\n",
       "      <td>48.741487</td>\n",
       "      <td>9.524736</td>\n",
       "      <td>5599.566000</td>\n",
       "      <td>4259.843100</td>\n",
       "      <td>5.635424</td>\n",
       "      <td>218.891600</td>\n",
       "      <td>1659.579315</td>\n",
       "      <td>18.531510</td>\n",
       "      <td>2075.835000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065006</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>113.933015</td>\n",
       "      <td>785.700975</td>\n",
       "      <td>15065.473670</td>\n",
       "      <td>5.566552</td>\n",
       "      <td>8.295481</td>\n",
       "      <td>122.778112</td>\n",
       "      <td>11.194395</td>\n",
       "      <td>...</td>\n",
       "      <td>1226.150310</td>\n",
       "      <td>111.363181</td>\n",
       "      <td>688.787250</td>\n",
       "      <td>2157.831000</td>\n",
       "      <td>1169.789050</td>\n",
       "      <td>460.672700</td>\n",
       "      <td>48.450610</td>\n",
       "      <td>1132.411695</td>\n",
       "      <td>87.753920</td>\n",
       "      <td>1949.938500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065007</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33.316627</td>\n",
       "      <td>2025.230581</td>\n",
       "      <td>11000.575797</td>\n",
       "      <td>6.985829</td>\n",
       "      <td>8.670795</td>\n",
       "      <td>9.668605</td>\n",
       "      <td>5.967618</td>\n",
       "      <td>...</td>\n",
       "      <td>369.325293</td>\n",
       "      <td>19.143244</td>\n",
       "      <td>418.350900</td>\n",
       "      <td>638.637600</td>\n",
       "      <td>558.308600</td>\n",
       "      <td>83.803710</td>\n",
       "      <td>55.825370</td>\n",
       "      <td>519.894006</td>\n",
       "      <td>376.097100</td>\n",
       "      <td>1048.790550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065008</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>72.755207</td>\n",
       "      <td>2160.700108</td>\n",
       "      <td>4732.540239</td>\n",
       "      <td>5.630027</td>\n",
       "      <td>13.312790</td>\n",
       "      <td>15.956931</td>\n",
       "      <td>18.454281</td>\n",
       "      <td>...</td>\n",
       "      <td>686.340507</td>\n",
       "      <td>26.663175</td>\n",
       "      <td>710.909300</td>\n",
       "      <td>1647.540000</td>\n",
       "      <td>736.967925</td>\n",
       "      <td>115.452500</td>\n",
       "      <td>55.752860</td>\n",
       "      <td>1529.392020</td>\n",
       "      <td>70.825420</td>\n",
       "      <td>1824.925000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065009</th>\n",
       "      <td>0.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>304.589916</td>\n",
       "      <td>177.046438</td>\n",
       "      <td>16235.837687</td>\n",
       "      <td>4.536025</td>\n",
       "      <td>39.009764</td>\n",
       "      <td>5.347365</td>\n",
       "      <td>14.975017</td>\n",
       "      <td>...</td>\n",
       "      <td>1186.185133</td>\n",
       "      <td>56.347652</td>\n",
       "      <td>7.243940</td>\n",
       "      <td>8044.482000</td>\n",
       "      <td>1204.401700</td>\n",
       "      <td>6.225645</td>\n",
       "      <td>555.139800</td>\n",
       "      <td>572.108800</td>\n",
       "      <td>7.657940</td>\n",
       "      <td>1744.423800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065010</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>88.871033</td>\n",
       "      <td>64.143112</td>\n",
       "      <td>13406.777035</td>\n",
       "      <td>5.555329</td>\n",
       "      <td>52.801878</td>\n",
       "      <td>6.534563</td>\n",
       "      <td>36.200404</td>\n",
       "      <td>...</td>\n",
       "      <td>290.075030</td>\n",
       "      <td>31.865941</td>\n",
       "      <td>82.448783</td>\n",
       "      <td>23655.970000</td>\n",
       "      <td>4542.206000</td>\n",
       "      <td>615.921700</td>\n",
       "      <td>9.425261</td>\n",
       "      <td>1533.098235</td>\n",
       "      <td>22.668210</td>\n",
       "      <td>31.769654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065011</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>142.123565</td>\n",
       "      <td>15.519496</td>\n",
       "      <td>695.019801</td>\n",
       "      <td>77.545878</td>\n",
       "      <td>9.972783</td>\n",
       "      <td>6.702547</td>\n",
       "      <td>9.548643</td>\n",
       "      <td>...</td>\n",
       "      <td>1984.753867</td>\n",
       "      <td>52.498025</td>\n",
       "      <td>1402.392150</td>\n",
       "      <td>1872.412000</td>\n",
       "      <td>440.552600</td>\n",
       "      <td>94.725430</td>\n",
       "      <td>758.315000</td>\n",
       "      <td>849.104085</td>\n",
       "      <td>2333.330000</td>\n",
       "      <td>1171.099250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065012</th>\n",
       "      <td>1.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>110.844892</td>\n",
       "      <td>17.575895</td>\n",
       "      <td>1473.933145</td>\n",
       "      <td>188.213344</td>\n",
       "      <td>8.538077</td>\n",
       "      <td>37.145432</td>\n",
       "      <td>15.087880</td>\n",
       "      <td>...</td>\n",
       "      <td>1877.260267</td>\n",
       "      <td>19.206398</td>\n",
       "      <td>2024.012350</td>\n",
       "      <td>2609.699000</td>\n",
       "      <td>609.782250</td>\n",
       "      <td>15.693150</td>\n",
       "      <td>847.401100</td>\n",
       "      <td>1005.177575</td>\n",
       "      <td>573.270900</td>\n",
       "      <td>1737.041050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065013</th>\n",
       "      <td>1.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>775.512251</td>\n",
       "      <td>74.360024</td>\n",
       "      <td>18.908550</td>\n",
       "      <td>6.635874</td>\n",
       "      <td>102.102438</td>\n",
       "      <td>9.471013</td>\n",
       "      <td>151.331233</td>\n",
       "      <td>...</td>\n",
       "      <td>1642.364717</td>\n",
       "      <td>250.260482</td>\n",
       "      <td>374.417150</td>\n",
       "      <td>113.096000</td>\n",
       "      <td>1396.942150</td>\n",
       "      <td>6.614659</td>\n",
       "      <td>2164.776000</td>\n",
       "      <td>3188.013055</td>\n",
       "      <td>8.046809</td>\n",
       "      <td>208.323260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065014</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>139.238129</td>\n",
       "      <td>12.278073</td>\n",
       "      <td>859.808859</td>\n",
       "      <td>20.916151</td>\n",
       "      <td>19.976150</td>\n",
       "      <td>28.014021</td>\n",
       "      <td>14.064752</td>\n",
       "      <td>...</td>\n",
       "      <td>2124.163200</td>\n",
       "      <td>64.509135</td>\n",
       "      <td>2756.464000</td>\n",
       "      <td>1880.746000</td>\n",
       "      <td>608.627350</td>\n",
       "      <td>286.626000</td>\n",
       "      <td>559.815500</td>\n",
       "      <td>1209.192700</td>\n",
       "      <td>1024.672000</td>\n",
       "      <td>1802.609350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065015</th>\n",
       "      <td>1.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>795.552432</td>\n",
       "      <td>269.115697</td>\n",
       "      <td>90.993129</td>\n",
       "      <td>10.004635</td>\n",
       "      <td>845.587370</td>\n",
       "      <td>9.121530</td>\n",
       "      <td>563.905454</td>\n",
       "      <td>...</td>\n",
       "      <td>1485.426500</td>\n",
       "      <td>125.628416</td>\n",
       "      <td>752.751346</td>\n",
       "      <td>6664.521000</td>\n",
       "      <td>3686.668450</td>\n",
       "      <td>27.448680</td>\n",
       "      <td>1879.931000</td>\n",
       "      <td>6399.885400</td>\n",
       "      <td>14.907730</td>\n",
       "      <td>40.532512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065016</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>301.731288</td>\n",
       "      <td>46.301725</td>\n",
       "      <td>1127.549298</td>\n",
       "      <td>15.824820</td>\n",
       "      <td>13.756153</td>\n",
       "      <td>93.186599</td>\n",
       "      <td>28.293208</td>\n",
       "      <td>...</td>\n",
       "      <td>3062.751667</td>\n",
       "      <td>92.283860</td>\n",
       "      <td>4569.766500</td>\n",
       "      <td>1447.083000</td>\n",
       "      <td>286.186950</td>\n",
       "      <td>454.789700</td>\n",
       "      <td>459.906400</td>\n",
       "      <td>1314.455420</td>\n",
       "      <td>1120.421000</td>\n",
       "      <td>1697.516650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065017</th>\n",
       "      <td>1.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>134.491817</td>\n",
       "      <td>28.629380</td>\n",
       "      <td>2293.169626</td>\n",
       "      <td>354.092338</td>\n",
       "      <td>16.930706</td>\n",
       "      <td>55.138294</td>\n",
       "      <td>21.293352</td>\n",
       "      <td>...</td>\n",
       "      <td>1760.789000</td>\n",
       "      <td>92.000255</td>\n",
       "      <td>1609.333250</td>\n",
       "      <td>2017.967000</td>\n",
       "      <td>782.346250</td>\n",
       "      <td>11.817800</td>\n",
       "      <td>132.724600</td>\n",
       "      <td>1429.673430</td>\n",
       "      <td>791.153000</td>\n",
       "      <td>1611.657100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065018</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.584943</td>\n",
       "      <td>10.491658</td>\n",
       "      <td>1351.294136</td>\n",
       "      <td>10.225083</td>\n",
       "      <td>34.094407</td>\n",
       "      <td>37.578248</td>\n",
       "      <td>14.314053</td>\n",
       "      <td>...</td>\n",
       "      <td>1562.662567</td>\n",
       "      <td>158.317573</td>\n",
       "      <td>1738.249750</td>\n",
       "      <td>875.925900</td>\n",
       "      <td>1846.119900</td>\n",
       "      <td>48.478250</td>\n",
       "      <td>152.610500</td>\n",
       "      <td>824.962060</td>\n",
       "      <td>2730.128000</td>\n",
       "      <td>1702.190500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065019</th>\n",
       "      <td>1.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>337.749114</td>\n",
       "      <td>27.918110</td>\n",
       "      <td>964.059163</td>\n",
       "      <td>10.245622</td>\n",
       "      <td>60.270189</td>\n",
       "      <td>13.446220</td>\n",
       "      <td>13.965202</td>\n",
       "      <td>...</td>\n",
       "      <td>2043.798567</td>\n",
       "      <td>218.581245</td>\n",
       "      <td>8.144527</td>\n",
       "      <td>21321.160000</td>\n",
       "      <td>260.603585</td>\n",
       "      <td>9.919398</td>\n",
       "      <td>2043.446000</td>\n",
       "      <td>2528.840565</td>\n",
       "      <td>134.158100</td>\n",
       "      <td>569.030284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065020</th>\n",
       "      <td>1.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.604577</td>\n",
       "      <td>11.283405</td>\n",
       "      <td>3702.603107</td>\n",
       "      <td>125.770351</td>\n",
       "      <td>8.426887</td>\n",
       "      <td>136.488644</td>\n",
       "      <td>13.957906</td>\n",
       "      <td>...</td>\n",
       "      <td>571.591107</td>\n",
       "      <td>45.289570</td>\n",
       "      <td>1529.138650</td>\n",
       "      <td>905.941900</td>\n",
       "      <td>223.114500</td>\n",
       "      <td>52.517560</td>\n",
       "      <td>25.286340</td>\n",
       "      <td>594.414595</td>\n",
       "      <td>958.443800</td>\n",
       "      <td>1532.995650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065021</th>\n",
       "      <td>1.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.742407</td>\n",
       "      <td>13.501060</td>\n",
       "      <td>7173.169584</td>\n",
       "      <td>75.255118</td>\n",
       "      <td>9.897242</td>\n",
       "      <td>39.587145</td>\n",
       "      <td>12.139993</td>\n",
       "      <td>...</td>\n",
       "      <td>538.261673</td>\n",
       "      <td>32.428083</td>\n",
       "      <td>667.700000</td>\n",
       "      <td>1490.194000</td>\n",
       "      <td>466.294455</td>\n",
       "      <td>19.938530</td>\n",
       "      <td>642.415700</td>\n",
       "      <td>571.269360</td>\n",
       "      <td>1197.677000</td>\n",
       "      <td>1026.412950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065022</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>234.644583</td>\n",
       "      <td>32.695949</td>\n",
       "      <td>416.943780</td>\n",
       "      <td>12.258123</td>\n",
       "      <td>13.743615</td>\n",
       "      <td>18.090270</td>\n",
       "      <td>19.668034</td>\n",
       "      <td>...</td>\n",
       "      <td>3233.478433</td>\n",
       "      <td>72.458528</td>\n",
       "      <td>2235.766000</td>\n",
       "      <td>1704.531000</td>\n",
       "      <td>313.049150</td>\n",
       "      <td>43.161960</td>\n",
       "      <td>173.647800</td>\n",
       "      <td>1114.012230</td>\n",
       "      <td>1685.934000</td>\n",
       "      <td>1945.261500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065023</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>123.994745</td>\n",
       "      <td>38.544932</td>\n",
       "      <td>938.150352</td>\n",
       "      <td>26.257133</td>\n",
       "      <td>10.681260</td>\n",
       "      <td>16.640757</td>\n",
       "      <td>11.363855</td>\n",
       "      <td>...</td>\n",
       "      <td>2336.463967</td>\n",
       "      <td>55.683422</td>\n",
       "      <td>1881.573500</td>\n",
       "      <td>1562.308000</td>\n",
       "      <td>229.795050</td>\n",
       "      <td>111.434200</td>\n",
       "      <td>446.037800</td>\n",
       "      <td>552.184325</td>\n",
       "      <td>1592.600000</td>\n",
       "      <td>1679.928100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065024</th>\n",
       "      <td>1.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>153.160805</td>\n",
       "      <td>8.016963</td>\n",
       "      <td>1356.889470</td>\n",
       "      <td>132.122170</td>\n",
       "      <td>15.087660</td>\n",
       "      <td>30.096220</td>\n",
       "      <td>25.897962</td>\n",
       "      <td>...</td>\n",
       "      <td>1820.072600</td>\n",
       "      <td>85.788764</td>\n",
       "      <td>3427.385000</td>\n",
       "      <td>1586.020000</td>\n",
       "      <td>439.842950</td>\n",
       "      <td>67.433840</td>\n",
       "      <td>286.176000</td>\n",
       "      <td>1012.676050</td>\n",
       "      <td>813.333800</td>\n",
       "      <td>1527.469000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065025</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>133.941268</td>\n",
       "      <td>21.845248</td>\n",
       "      <td>896.246781</td>\n",
       "      <td>76.819437</td>\n",
       "      <td>13.285551</td>\n",
       "      <td>10.420094</td>\n",
       "      <td>9.449720</td>\n",
       "      <td>...</td>\n",
       "      <td>2650.085233</td>\n",
       "      <td>67.113705</td>\n",
       "      <td>1859.606000</td>\n",
       "      <td>1083.613000</td>\n",
       "      <td>329.333600</td>\n",
       "      <td>73.667470</td>\n",
       "      <td>206.264900</td>\n",
       "      <td>707.231965</td>\n",
       "      <td>1925.581000</td>\n",
       "      <td>1548.703850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065026</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>499.330427</td>\n",
       "      <td>15.351979</td>\n",
       "      <td>478.637090</td>\n",
       "      <td>12.008191</td>\n",
       "      <td>10.749362</td>\n",
       "      <td>12.895410</td>\n",
       "      <td>11.512859</td>\n",
       "      <td>...</td>\n",
       "      <td>4729.760600</td>\n",
       "      <td>76.104755</td>\n",
       "      <td>3139.699500</td>\n",
       "      <td>6694.573000</td>\n",
       "      <td>1845.633400</td>\n",
       "      <td>713.101300</td>\n",
       "      <td>562.988800</td>\n",
       "      <td>2239.454265</td>\n",
       "      <td>467.278900</td>\n",
       "      <td>2250.084000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065027</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>436.495579</td>\n",
       "      <td>28.155839</td>\n",
       "      <td>501.557758</td>\n",
       "      <td>1328.846867</td>\n",
       "      <td>127.293009</td>\n",
       "      <td>164.394109</td>\n",
       "      <td>14.470248</td>\n",
       "      <td>...</td>\n",
       "      <td>2343.621453</td>\n",
       "      <td>127.447265</td>\n",
       "      <td>4752.332000</td>\n",
       "      <td>687.311800</td>\n",
       "      <td>93.791930</td>\n",
       "      <td>1492.457000</td>\n",
       "      <td>241.484800</td>\n",
       "      <td>2063.848065</td>\n",
       "      <td>535.111900</td>\n",
       "      <td>1718.573100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065028</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.558639</td>\n",
       "      <td>740.139794</td>\n",
       "      <td>8290.655792</td>\n",
       "      <td>88.716041</td>\n",
       "      <td>15.725767</td>\n",
       "      <td>10.029241</td>\n",
       "      <td>471.003930</td>\n",
       "      <td>...</td>\n",
       "      <td>17.967367</td>\n",
       "      <td>425.330310</td>\n",
       "      <td>79.570540</td>\n",
       "      <td>197.393600</td>\n",
       "      <td>667.582110</td>\n",
       "      <td>42.838550</td>\n",
       "      <td>11.421720</td>\n",
       "      <td>676.320480</td>\n",
       "      <td>9.553312</td>\n",
       "      <td>2363.307900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM4065029</th>\n",
       "      <td>1.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>392.801435</td>\n",
       "      <td>22.193727</td>\n",
       "      <td>1888.120143</td>\n",
       "      <td>338.872848</td>\n",
       "      <td>36.854081</td>\n",
       "      <td>117.416402</td>\n",
       "      <td>60.488695</td>\n",
       "      <td>...</td>\n",
       "      <td>1574.187453</td>\n",
       "      <td>67.244330</td>\n",
       "      <td>2713.013000</td>\n",
       "      <td>448.175500</td>\n",
       "      <td>211.614450</td>\n",
       "      <td>8.809586</td>\n",
       "      <td>125.637600</td>\n",
       "      <td>1160.820610</td>\n",
       "      <td>1045.120000</td>\n",
       "      <td>1952.216350</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60 rows × 16991 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Brugada Syndrome    Age  Gender         A1BG         A1CF  \\\n",
       "GSM4064970               0.0   24.0     1.0    62.355348    14.440420   \n",
       "GSM4064971               0.0  112.0     1.0    10.278570     7.569081   \n",
       "GSM4064972               0.0    8.0     0.0    54.362789     5.384082   \n",
       "GSM4064973               0.0   24.0     1.0     5.764988     8.260571   \n",
       "GSM4064974               0.0    0.6     0.0    23.992323    18.493197   \n",
       "GSM4064975               0.0   72.0     1.0    31.366236     5.428562   \n",
       "GSM4064976               0.0   24.0     0.0   267.552914     8.712279   \n",
       "GSM4064977               0.0   36.0     1.0  1879.845723   287.377749   \n",
       "GSM4064978               0.0   52.0     1.0   112.182100     7.417203   \n",
       "GSM4064979               0.0   20.0     1.0    78.592842    30.706918   \n",
       "GSM4064980               0.0   24.0     0.0     9.500891     8.117416   \n",
       "GSM4064981               0.0    0.0     0.0     6.975764     7.315140   \n",
       "GSM4064982               0.0    0.0     0.0    31.740504     9.078905   \n",
       "GSM4064983               0.0   80.0     0.0   236.178258    38.379484   \n",
       "GSM4064984               0.0   52.0     0.0     7.733671    15.222632   \n",
       "GSM4064985               0.0    0.5     0.0    11.728377     8.154470   \n",
       "GSM4064986               0.0  144.0     1.0    11.679665     7.576425   \n",
       "GSM4064987               0.0    0.0     1.0    47.697543   151.769953   \n",
       "GSM4064988               0.0   24.0     1.0    75.198431     6.857917   \n",
       "GSM4064989               0.0    0.0     1.0   586.805849   111.214584   \n",
       "GSM4064990               0.0  112.0     1.0   234.311653  5948.300181   \n",
       "GSM4064991               0.0   24.0     1.0    29.966833  1459.044789   \n",
       "GSM4064992               0.0    8.0     0.0   142.618631  2712.974607   \n",
       "GSM4064993               0.0    0.6     0.0    26.504939  1482.934434   \n",
       "GSM4064994               0.0   20.0     1.0    10.936492  3833.436124   \n",
       "GSM4064995               0.0   36.0     1.0   726.484206    12.763040   \n",
       "GSM4064996               0.0   12.0     0.0  2159.771081    15.451472   \n",
       "GSM4064997               0.0   72.0     1.0    38.209843   636.232491   \n",
       "GSM4064998               0.0    0.0     0.0   213.324275  1325.104073   \n",
       "GSM4064999               0.0   52.0     1.0     8.643003   854.772741   \n",
       "GSM4065000               0.0   24.0     1.0   115.095751  1013.827503   \n",
       "GSM4065001               0.0    0.0     0.0    48.668886  5410.263027   \n",
       "GSM4065002               0.0    0.0     0.0    26.398669   871.941258   \n",
       "GSM4065003               0.0    0.0     0.0    28.387492  3399.569849   \n",
       "GSM4065004               0.0  144.0     1.0   784.652359  2326.449789   \n",
       "GSM4065005               0.0   52.0     0.0  1600.291058  1896.662925   \n",
       "GSM4065006               0.0    0.0     0.0   113.933015   785.700975   \n",
       "GSM4065007               0.0   24.0     1.0    33.316627  2025.230581   \n",
       "GSM4065008               0.0    2.0     1.0    72.755207  2160.700108   \n",
       "GSM4065009               0.0   80.0     0.0   304.589916   177.046438   \n",
       "GSM4065010               0.0   24.0     0.0    88.871033    64.143112   \n",
       "GSM4065011               1.0    8.0     0.0   142.123565    15.519496   \n",
       "GSM4065012               1.0   20.0     1.0   110.844892    17.575895   \n",
       "GSM4065013               1.0   24.0     1.0   775.512251    74.360024   \n",
       "GSM4065014               1.0    0.6     0.0   139.238129    12.278073   \n",
       "GSM4065015               1.0   72.0     1.0   795.552432   269.115697   \n",
       "GSM4065016               1.0    0.0     0.0   301.731288    46.301725   \n",
       "GSM4065017               1.0   36.0     1.0   134.491817    28.629380   \n",
       "GSM4065018               1.0    0.0     1.0    34.584943    10.491658   \n",
       "GSM4065019               1.0   24.0     1.0   337.749114    27.918110   \n",
       "GSM4065020               1.0   60.0     0.0    68.604577    11.283405   \n",
       "GSM4065021               1.0   52.0     0.0    33.742407    13.501060   \n",
       "GSM4065022               1.0    0.0     0.0   234.644583    32.695949   \n",
       "GSM4065023               1.0    0.0     0.0   123.994745    38.544932   \n",
       "GSM4065024               1.0   52.0     0.0   153.160805     8.016963   \n",
       "GSM4065025               1.0    0.0     1.0   133.941268    21.845248   \n",
       "GSM4065026               1.0    0.0     0.0   499.330427    15.351979   \n",
       "GSM4065027               1.0    2.0     1.0   436.495579    28.155839   \n",
       "GSM4065028               1.0    2.0     0.0    59.558639   740.139794   \n",
       "GSM4065029               1.0  144.0     1.0   392.801435    22.193727   \n",
       "\n",
       "                     A2M        A2ML1      A3GALT2       A4GALT        A4GNT  \\\n",
       "GSM4064970   2460.196694     4.686300     6.105481   587.097412    16.382563   \n",
       "GSM4064971   7293.759251     5.559912     4.979804   327.278673     6.757537   \n",
       "GSM4064972   1536.235125   195.513335    65.989054  2739.802065   109.323537   \n",
       "GSM4064973   4517.352275     5.725376     6.391028   806.762413    45.303023   \n",
       "GSM4064974   4215.080680    17.109983     7.504535   923.792662     9.095975   \n",
       "GSM4064975   1180.786270     5.852868    22.394404   958.550157    73.304113   \n",
       "GSM4064976   8764.367836    37.777003   438.316512     4.904852     6.736836   \n",
       "GSM4064977      7.111364     5.390598   245.963630    80.346495  1189.334982   \n",
       "GSM4064978   3260.866611     7.106071    10.197472   268.349773    17.072828   \n",
       "GSM4064979   1559.723056     9.450259    10.040594   611.063243    74.828051   \n",
       "GSM4064980   6556.304666     3.985734     7.040947   226.164405    11.574808   \n",
       "GSM4064981   3293.237621     6.412876     5.016920   495.804636     8.940770   \n",
       "GSM4064982   3562.690498    25.182167     9.602690   148.658069    12.069092   \n",
       "GSM4064983   5261.985908     4.361902   174.406763  2192.210826   231.662768   \n",
       "GSM4064984   3034.285917     8.050865     8.015696   820.683189    14.982735   \n",
       "GSM4064985   8701.918648     4.693914     4.645250   575.679296     9.831937   \n",
       "GSM4064986   5548.304952     9.535022     4.319101   757.620403    11.472898   \n",
       "GSM4064987   4480.836105     4.439174     6.913088   857.596072    51.583578   \n",
       "GSM4064988   2615.503687     7.667636     7.647184   180.074405     8.011405   \n",
       "GSM4064989   1404.553002    63.813658     5.573835     6.926686     8.391545   \n",
       "GSM4064990  14769.609246     8.206099     5.786520     9.875059    16.007956   \n",
       "GSM4064991   9684.632839     5.403382     5.623221    99.371524    23.985891   \n",
       "GSM4064992   9394.966901     5.294361     6.947625    21.549403     8.514557   \n",
       "GSM4064993  11401.858638     4.534517     6.299148   122.005959     9.861370   \n",
       "GSM4064994  13289.794414     4.225422     5.420832    25.827550     6.355564   \n",
       "GSM4064995   5514.580124   166.227748     9.098739     4.948842   561.678828   \n",
       "GSM4064996   3746.966295    19.563223    52.677832   243.517783    27.023779   \n",
       "GSM4064997  10756.733813     5.291773     7.199933   135.180297     6.741027   \n",
       "GSM4064998  12237.567394     5.526838     5.766445    25.970225    22.744779   \n",
       "GSM4064999      5.534308    15.926672  2955.620292     5.534308   222.287611   \n",
       "GSM4065000   7876.024509     8.166187    17.061461   257.494382    80.656827   \n",
       "GSM4065001  15126.746543     4.398537     7.060038     8.176693    10.599370   \n",
       "GSM4065002   8889.129204     8.516182     6.951950   834.090544    14.354971   \n",
       "GSM4065003  14975.615481     5.865120     9.070901    16.649167     5.892301   \n",
       "GSM4065004   8997.131522     6.664390     9.163141     8.417904    10.878410   \n",
       "GSM4065005   8895.637507    18.668081     6.935069    49.675526   352.678070   \n",
       "GSM4065006  15065.473670     5.566552     8.295481   122.778112    11.194395   \n",
       "GSM4065007  11000.575797     6.985829     8.670795     9.668605     5.967618   \n",
       "GSM4065008   4732.540239     5.630027    13.312790    15.956931    18.454281   \n",
       "GSM4065009  16235.837687     4.536025    39.009764     5.347365    14.975017   \n",
       "GSM4065010  13406.777035     5.555329    52.801878     6.534563    36.200404   \n",
       "GSM4065011    695.019801    77.545878     9.972783     6.702547     9.548643   \n",
       "GSM4065012   1473.933145   188.213344     8.538077    37.145432    15.087880   \n",
       "GSM4065013     18.908550     6.635874   102.102438     9.471013   151.331233   \n",
       "GSM4065014    859.808859    20.916151    19.976150    28.014021    14.064752   \n",
       "GSM4065015     90.993129    10.004635   845.587370     9.121530   563.905454   \n",
       "GSM4065016   1127.549298    15.824820    13.756153    93.186599    28.293208   \n",
       "GSM4065017   2293.169626   354.092338    16.930706    55.138294    21.293352   \n",
       "GSM4065018   1351.294136    10.225083    34.094407    37.578248    14.314053   \n",
       "GSM4065019    964.059163    10.245622    60.270189    13.446220    13.965202   \n",
       "GSM4065020   3702.603107   125.770351     8.426887   136.488644    13.957906   \n",
       "GSM4065021   7173.169584    75.255118     9.897242    39.587145    12.139993   \n",
       "GSM4065022    416.943780    12.258123    13.743615    18.090270    19.668034   \n",
       "GSM4065023    938.150352    26.257133    10.681260    16.640757    11.363855   \n",
       "GSM4065024   1356.889470   132.122170    15.087660    30.096220    25.897962   \n",
       "GSM4065025    896.246781    76.819437    13.285551    10.420094     9.449720   \n",
       "GSM4065026    478.637090    12.008191    10.749362    12.895410    11.512859   \n",
       "GSM4065027    501.557758  1328.846867   127.293009   164.394109    14.470248   \n",
       "GSM4065028   8290.655792    88.716041    15.725767    10.029241   471.003930   \n",
       "GSM4065029   1888.120143   338.872848    36.854081   117.416402    60.488695   \n",
       "\n",
       "            ...       ZWILCH        ZWINT         ZXDA          ZXDB  \\\n",
       "GSM4064970  ...   475.836640    10.585559  1858.563000    309.923000   \n",
       "GSM4064971  ...   114.519993    10.859937   452.635850    432.893700   \n",
       "GSM4064972  ...   907.482433    25.361408  2418.226000      6.477754   \n",
       "GSM4064973  ...   492.995270    13.643631   568.066500   3557.904000   \n",
       "GSM4064974  ...   449.435800    13.533547  1767.799450    235.760200   \n",
       "GSM4064975  ...   901.280519    16.117974  2119.380000    762.076200   \n",
       "GSM4064976  ...   231.004713    11.046941   401.668940   7046.628000   \n",
       "GSM4064977  ...   320.578720    10.155646   733.606260      7.825502   \n",
       "GSM4064978  ...   369.459033     8.997515   181.475945  11267.170000   \n",
       "GSM4064979  ...   644.490500    11.755870  2287.628500     26.475930   \n",
       "GSM4064980  ...   143.642440    10.698736   213.605250    606.987800   \n",
       "GSM4064981  ...   489.470387    60.020860   895.049800    723.229200   \n",
       "GSM4064982  ...   291.751847    28.721025   646.298115   5751.135000   \n",
       "GSM4064983  ...    42.409687    23.656955  2197.597702    360.965800   \n",
       "GSM4064984  ...   828.796156   128.576423  1711.594000     13.635600   \n",
       "GSM4064985  ...   536.601530   111.695540   555.470200    937.922200   \n",
       "GSM4064986  ...   512.308813    18.318939  1472.623500    820.456000   \n",
       "GSM4064987  ...   529.796164    27.621451   778.790050   1150.408000   \n",
       "GSM4064988  ...   170.617420    80.205718    19.971890  12410.470000   \n",
       "GSM4064989  ...   231.132710    35.340658     5.441949     11.570360   \n",
       "GSM4064990  ...   186.685233     9.615675   352.192000   1318.780000   \n",
       "GSM4064991  ...   648.563097    28.267225   292.516800    856.175300   \n",
       "GSM4064992  ...   943.467097    64.065413  1153.041450   1286.365000   \n",
       "GSM4064993  ...   563.999787    58.405770   683.971150   1100.959000   \n",
       "GSM4064994  ...   280.885937     9.353346   306.606800    974.010300   \n",
       "GSM4064995  ...    30.660071    36.665401   132.786752     73.986010   \n",
       "GSM4064996  ...   183.798180   119.650187   128.729310      6.084738   \n",
       "GSM4064997  ...   497.643930    19.221363   333.777700   3602.606000   \n",
       "GSM4064998  ...   840.886717   134.264273   419.692650    223.958500   \n",
       "GSM4064999  ...  3575.796034  2550.317877   588.894001    172.472900   \n",
       "GSM4065000  ...   753.020108    15.955018   455.715100   2080.280000   \n",
       "GSM4065001  ...   477.573507   175.482793   163.191750    365.437300   \n",
       "GSM4065002  ...  1132.310457   219.915290   649.928000   2584.695000   \n",
       "GSM4065003  ...   403.493293    92.251820   234.376500    835.752100   \n",
       "GSM4065004  ...   410.482363    14.841379   697.120800   3235.150000   \n",
       "GSM4065005  ...    93.483760    48.741487     9.524736   5599.566000   \n",
       "GSM4065006  ...  1226.150310   111.363181   688.787250   2157.831000   \n",
       "GSM4065007  ...   369.325293    19.143244   418.350900    638.637600   \n",
       "GSM4065008  ...   686.340507    26.663175   710.909300   1647.540000   \n",
       "GSM4065009  ...  1186.185133    56.347652     7.243940   8044.482000   \n",
       "GSM4065010  ...   290.075030    31.865941    82.448783  23655.970000   \n",
       "GSM4065011  ...  1984.753867    52.498025  1402.392150   1872.412000   \n",
       "GSM4065012  ...  1877.260267    19.206398  2024.012350   2609.699000   \n",
       "GSM4065013  ...  1642.364717   250.260482   374.417150    113.096000   \n",
       "GSM4065014  ...  2124.163200    64.509135  2756.464000   1880.746000   \n",
       "GSM4065015  ...  1485.426500   125.628416   752.751346   6664.521000   \n",
       "GSM4065016  ...  3062.751667    92.283860  4569.766500   1447.083000   \n",
       "GSM4065017  ...  1760.789000    92.000255  1609.333250   2017.967000   \n",
       "GSM4065018  ...  1562.662567   158.317573  1738.249750    875.925900   \n",
       "GSM4065019  ...  2043.798567   218.581245     8.144527  21321.160000   \n",
       "GSM4065020  ...   571.591107    45.289570  1529.138650    905.941900   \n",
       "GSM4065021  ...   538.261673    32.428083   667.700000   1490.194000   \n",
       "GSM4065022  ...  3233.478433    72.458528  2235.766000   1704.531000   \n",
       "GSM4065023  ...  2336.463967    55.683422  1881.573500   1562.308000   \n",
       "GSM4065024  ...  1820.072600    85.788764  3427.385000   1586.020000   \n",
       "GSM4065025  ...  2650.085233    67.113705  1859.606000   1083.613000   \n",
       "GSM4065026  ...  4729.760600    76.104755  3139.699500   6694.573000   \n",
       "GSM4065027  ...  2343.621453   127.447265  4752.332000    687.311800   \n",
       "GSM4065028  ...    17.967367   425.330310    79.570540    197.393600   \n",
       "GSM4065029  ...  1574.187453    67.244330  2713.013000    448.175500   \n",
       "\n",
       "                   ZXDC       ZYG11A       ZYG11B          ZYX        ZZEF1  \\\n",
       "GSM4064970   238.824645    16.280420    27.983100   925.877535   457.178400   \n",
       "GSM4064971   342.556525     6.001372   328.690900   323.831265  1071.818000   \n",
       "GSM4064972     6.872454    10.014170   182.508200  1914.635812     6.965921   \n",
       "GSM4064973   933.398150     7.851095   308.473100  1901.632360   165.038800   \n",
       "GSM4064974   122.383625    45.240360    30.014700   901.722045   352.542300   \n",
       "GSM4064975   391.236405     9.205301     4.563783  2633.665940     5.055491   \n",
       "GSM4064976   244.577066     3.651659  2042.331000  1020.103037   162.418400   \n",
       "GSM4064977    93.016329   219.472800     7.363958  8311.804700     4.340990   \n",
       "GSM4064978  1351.171000     7.917643   126.146200  3719.357300   241.948900   \n",
       "GSM4064979   189.163285    12.040690    18.647120  1397.042660     4.680267   \n",
       "GSM4064980   294.037245     8.095774   164.823500   476.196355   971.513800   \n",
       "GSM4064981   474.962190    15.774970   204.312500   622.720075   366.868200   \n",
       "GSM4064982  1758.740500    25.433950   200.265300   537.587265   278.035600   \n",
       "GSM4064983    16.988810   801.090400    63.027030  2246.607310     5.163648   \n",
       "GSM4064984    64.651605     5.867572    47.443430  1135.455990     6.997004   \n",
       "GSM4064985   469.344995     8.042021   187.039300   561.487255   690.213200   \n",
       "GSM4064986   583.374900     4.111516   158.076200   704.123905   254.941600   \n",
       "GSM4064987   293.628510   149.278600     8.218412  1373.463340   462.977200   \n",
       "GSM4064988  1020.046650     8.103821   457.943000  1123.434000   262.584300   \n",
       "GSM4064989  1434.869861    10.258190  1059.371000  5155.736470     4.640261   \n",
       "GSM4064990  2260.991300   399.729500   183.506900   142.258435  1559.518000   \n",
       "GSM4064991   478.450515    67.469250    40.998740   492.073800   545.807800   \n",
       "GSM4064992   634.685975   472.844100   148.258000   834.830370   384.734100   \n",
       "GSM4064993   374.124685   251.940300    90.164050   811.157355   318.022700   \n",
       "GSM4064994   515.183950    79.411110   116.111700   281.953865   662.150700   \n",
       "GSM4064995  2135.142185   125.304800   943.686500   973.182250    10.223680   \n",
       "GSM4064996   119.029350   432.274800   673.329300  5206.168050   179.628800   \n",
       "GSM4064997  1090.534000    46.158280    50.135550   519.852540   887.203800   \n",
       "GSM4064998   487.167450   458.851300    43.845360   528.930775     6.546703   \n",
       "GSM4064999     6.128139  5045.076000   264.535500    11.737111  5999.011000   \n",
       "GSM4065000  1598.767650   907.107300     6.655668  1513.037035   492.228700   \n",
       "GSM4065001  1327.685945   140.720900   134.320000   319.536445   795.299100   \n",
       "GSM4065002   514.736750   143.806900    61.196830  1258.490470   412.119300   \n",
       "GSM4065003   493.296640   162.268600    74.689590   320.661088   497.860700   \n",
       "GSM4065004  2570.936850     5.946879   157.209600  1014.134508     4.868951   \n",
       "GSM4065005  4259.843100     5.635424   218.891600  1659.579315    18.531510   \n",
       "GSM4065006  1169.789050   460.672700    48.450610  1132.411695    87.753920   \n",
       "GSM4065007   558.308600    83.803710    55.825370   519.894006   376.097100   \n",
       "GSM4065008   736.967925   115.452500    55.752860  1529.392020    70.825420   \n",
       "GSM4065009  1204.401700     6.225645   555.139800   572.108800     7.657940   \n",
       "GSM4065010  4542.206000   615.921700     9.425261  1533.098235    22.668210   \n",
       "GSM4065011   440.552600    94.725430   758.315000   849.104085  2333.330000   \n",
       "GSM4065012   609.782250    15.693150   847.401100  1005.177575   573.270900   \n",
       "GSM4065013  1396.942150     6.614659  2164.776000  3188.013055     8.046809   \n",
       "GSM4065014   608.627350   286.626000   559.815500  1209.192700  1024.672000   \n",
       "GSM4065015  3686.668450    27.448680  1879.931000  6399.885400    14.907730   \n",
       "GSM4065016   286.186950   454.789700   459.906400  1314.455420  1120.421000   \n",
       "GSM4065017   782.346250    11.817800   132.724600  1429.673430   791.153000   \n",
       "GSM4065018  1846.119900    48.478250   152.610500   824.962060  2730.128000   \n",
       "GSM4065019   260.603585     9.919398  2043.446000  2528.840565   134.158100   \n",
       "GSM4065020   223.114500    52.517560    25.286340   594.414595   958.443800   \n",
       "GSM4065021   466.294455    19.938530   642.415700   571.269360  1197.677000   \n",
       "GSM4065022   313.049150    43.161960   173.647800  1114.012230  1685.934000   \n",
       "GSM4065023   229.795050   111.434200   446.037800   552.184325  1592.600000   \n",
       "GSM4065024   439.842950    67.433840   286.176000  1012.676050   813.333800   \n",
       "GSM4065025   329.333600    73.667470   206.264900   707.231965  1925.581000   \n",
       "GSM4065026  1845.633400   713.101300   562.988800  2239.454265   467.278900   \n",
       "GSM4065027    93.791930  1492.457000   241.484800  2063.848065   535.111900   \n",
       "GSM4065028   667.582110    42.838550    11.421720   676.320480     9.553312   \n",
       "GSM4065029   211.614450     8.809586   125.637600  1160.820610  1045.120000   \n",
       "\n",
       "                   ZZZ3  \n",
       "GSM4064970  1711.636700  \n",
       "GSM4064971  1576.779250  \n",
       "GSM4064972  1755.359300  \n",
       "GSM4064973  1326.673000  \n",
       "GSM4064974  1327.331950  \n",
       "GSM4064975   926.780490  \n",
       "GSM4064976     8.467687  \n",
       "GSM4064977   169.654780  \n",
       "GSM4064978   887.739150  \n",
       "GSM4064979  1564.791560  \n",
       "GSM4064980   669.199000  \n",
       "GSM4064981  1844.402700  \n",
       "GSM4064982   375.949015  \n",
       "GSM4064983  2987.136878  \n",
       "GSM4064984   729.671550  \n",
       "GSM4064985  1559.428500  \n",
       "GSM4064986  1842.112400  \n",
       "GSM4064987  1209.270365  \n",
       "GSM4064988   359.373570  \n",
       "GSM4064989     6.326132  \n",
       "GSM4064990  1488.380950  \n",
       "GSM4064991  2002.681900  \n",
       "GSM4064992  2447.874200  \n",
       "GSM4064993  1691.558100  \n",
       "GSM4064994  1502.183100  \n",
       "GSM4064995    38.197745  \n",
       "GSM4064996   380.590205  \n",
       "GSM4064997  1722.465000  \n",
       "GSM4064998  1856.440850  \n",
       "GSM4064999   149.815400  \n",
       "GSM4065000  1258.421950  \n",
       "GSM4065001  1165.559900  \n",
       "GSM4065002  1642.838750  \n",
       "GSM4065003  1677.421700  \n",
       "GSM4065004  1663.368800  \n",
       "GSM4065005  2075.835000  \n",
       "GSM4065006  1949.938500  \n",
       "GSM4065007  1048.790550  \n",
       "GSM4065008  1824.925000  \n",
       "GSM4065009  1744.423800  \n",
       "GSM4065010    31.769654  \n",
       "GSM4065011  1171.099250  \n",
       "GSM4065012  1737.041050  \n",
       "GSM4065013   208.323260  \n",
       "GSM4065014  1802.609350  \n",
       "GSM4065015    40.532512  \n",
       "GSM4065016  1697.516650  \n",
       "GSM4065017  1611.657100  \n",
       "GSM4065018  1702.190500  \n",
       "GSM4065019   569.030284  \n",
       "GSM4065020  1532.995650  \n",
       "GSM4065021  1026.412950  \n",
       "GSM4065022  1945.261500  \n",
       "GSM4065023  1679.928100  \n",
       "GSM4065024  1527.469000  \n",
       "GSM4065025  1548.703850  \n",
       "GSM4065026  2250.084000  \n",
       "GSM4065027  1718.573100  \n",
       "GSM4065028  2363.307900  \n",
       "GSM4065029  1952.216350  \n",
       "\n",
       "[60 rows x 16991 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:35.158646754Z",
     "start_time": "2024-01-02T01:03:35.156975304Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Brugada Syndrome', the least common label is '1.0' with 19 occurrences. This represents 31.67% of the dataset.\n",
      "The distribution of the feature 'Brugada Syndrome' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 0.375\n",
      "  50% (Median): 24.0\n",
      "  75%: 52.0\n",
      "Min: 0.0\n",
      "Max: 144.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 30 occurrences. This represents 50.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:35.158803015Z",
     "start_time": "2024-01-02T01:03:35.157063388Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: ../../output1/Yuxuan/Brugada-Syndrome/cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.004312476Z",
     "start_time": "2024-01-02T01:03:35.157165158Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)\n",
    "#merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.048788605Z",
     "start_time": "2024-01-02T01:03:36.005925887Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.049183603Z",
     "start_time": "2024-01-02T01:03:36.048440305Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSE136992</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cohort_id  is_usable  is_available  is_biased  has_age  has_gender  \\\n",
       "0  GSE136992       True          True      False     True        True   \n",
       "\n",
       "   sample_size note  \n",
       "0           60       "
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH, 'has_age')\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GSE136992'"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.345375576Z",
     "start_time": "2024-01-02T01:03:36.049289340Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Brugada Syndrome</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62.355348</td>\n",
       "      <td>14.440420</td>\n",
       "      <td>2460.196694</td>\n",
       "      <td>4.686300</td>\n",
       "      <td>6.105481</td>\n",
       "      <td>587.097412</td>\n",
       "      <td>16.382563</td>\n",
       "      <td>...</td>\n",
       "      <td>475.836640</td>\n",
       "      <td>10.585558</td>\n",
       "      <td>1858.56300</td>\n",
       "      <td>309.923000</td>\n",
       "      <td>238.824645</td>\n",
       "      <td>16.280420</td>\n",
       "      <td>27.9831</td>\n",
       "      <td>925.877535</td>\n",
       "      <td>457.178400</td>\n",
       "      <td>1711.63670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.278570</td>\n",
       "      <td>7.569081</td>\n",
       "      <td>7293.759251</td>\n",
       "      <td>5.559912</td>\n",
       "      <td>4.979804</td>\n",
       "      <td>327.278673</td>\n",
       "      <td>6.757537</td>\n",
       "      <td>...</td>\n",
       "      <td>114.519993</td>\n",
       "      <td>10.859937</td>\n",
       "      <td>452.63585</td>\n",
       "      <td>432.893700</td>\n",
       "      <td>342.556525</td>\n",
       "      <td>6.001372</td>\n",
       "      <td>328.6909</td>\n",
       "      <td>323.831265</td>\n",
       "      <td>1071.818000</td>\n",
       "      <td>1576.77925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54.362789</td>\n",
       "      <td>5.384082</td>\n",
       "      <td>1536.235125</td>\n",
       "      <td>195.513335</td>\n",
       "      <td>65.989054</td>\n",
       "      <td>2739.802065</td>\n",
       "      <td>109.323537</td>\n",
       "      <td>...</td>\n",
       "      <td>907.482433</td>\n",
       "      <td>25.361408</td>\n",
       "      <td>2418.22600</td>\n",
       "      <td>6.477754</td>\n",
       "      <td>6.872454</td>\n",
       "      <td>10.014170</td>\n",
       "      <td>182.5082</td>\n",
       "      <td>1914.635812</td>\n",
       "      <td>6.965921</td>\n",
       "      <td>1755.35930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.764988</td>\n",
       "      <td>8.260571</td>\n",
       "      <td>4517.352275</td>\n",
       "      <td>5.725376</td>\n",
       "      <td>6.391028</td>\n",
       "      <td>806.762413</td>\n",
       "      <td>45.303023</td>\n",
       "      <td>...</td>\n",
       "      <td>492.995270</td>\n",
       "      <td>13.643631</td>\n",
       "      <td>568.06650</td>\n",
       "      <td>3557.904000</td>\n",
       "      <td>933.398150</td>\n",
       "      <td>7.851095</td>\n",
       "      <td>308.4731</td>\n",
       "      <td>1901.632360</td>\n",
       "      <td>165.038800</td>\n",
       "      <td>1326.67300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.992323</td>\n",
       "      <td>18.493197</td>\n",
       "      <td>4215.080680</td>\n",
       "      <td>17.109983</td>\n",
       "      <td>7.504535</td>\n",
       "      <td>923.792662</td>\n",
       "      <td>9.095975</td>\n",
       "      <td>...</td>\n",
       "      <td>449.435800</td>\n",
       "      <td>13.533547</td>\n",
       "      <td>1767.79945</td>\n",
       "      <td>235.760200</td>\n",
       "      <td>122.383625</td>\n",
       "      <td>45.240360</td>\n",
       "      <td>30.0147</td>\n",
       "      <td>901.722045</td>\n",
       "      <td>352.542300</td>\n",
       "      <td>1327.33195</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 16991 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Brugada Syndrome    Age  Gender       A1BG       A1CF          A2M  \\\n",
       "0               0.0   24.0     1.0  62.355348  14.440420  2460.196694   \n",
       "1               0.0  112.0     1.0  10.278570   7.569081  7293.759251   \n",
       "2               0.0    8.0     0.0  54.362789   5.384082  1536.235125   \n",
       "3               0.0   24.0     1.0   5.764988   8.260571  4517.352275   \n",
       "4               0.0    0.6     0.0  23.992323  18.493197  4215.080680   \n",
       "\n",
       "        A2ML1    A3GALT2       A4GALT       A4GNT  ...      ZWILCH      ZWINT  \\\n",
       "0    4.686300   6.105481   587.097412   16.382563  ...  475.836640  10.585558   \n",
       "1    5.559912   4.979804   327.278673    6.757537  ...  114.519993  10.859937   \n",
       "2  195.513335  65.989054  2739.802065  109.323537  ...  907.482433  25.361408   \n",
       "3    5.725376   6.391028   806.762413   45.303023  ...  492.995270  13.643631   \n",
       "4   17.109983   7.504535   923.792662    9.095975  ...  449.435800  13.533547   \n",
       "\n",
       "         ZXDA         ZXDB        ZXDC     ZYG11A    ZYG11B          ZYX  \\\n",
       "0  1858.56300   309.923000  238.824645  16.280420   27.9831   925.877535   \n",
       "1   452.63585   432.893700  342.556525   6.001372  328.6909   323.831265   \n",
       "2  2418.22600     6.477754    6.872454  10.014170  182.5082  1914.635812   \n",
       "3   568.06650  3557.904000  933.398150   7.851095  308.4731  1901.632360   \n",
       "4  1767.79945   235.760200  122.383625  45.240360   30.0147   901.722045   \n",
       "\n",
       "         ZZEF1        ZZZ3  \n",
       "0   457.178400  1711.63670  \n",
       "1  1071.818000  1576.77925  \n",
       "2     6.965921  1755.35930  \n",
       "3   165.038800  1326.67300  \n",
       "4   352.542300  1327.33195  \n",
       "\n",
       "[5 rows x 16991 columns]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, condition_best_cohort + '.csv'))\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.401002087Z",
     "start_time": "2024-01-02T01:03:36.400659608Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# If both age and gender features are available, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "# Remove the other condition to prevent interference.\n",
    "merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.403035823Z",
     "start_time": "2024-01-02T01:03:36.401081946Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "X = merged_data.drop(columns=[TRAIT, condition]).values\n",
    "Y = merged_data[TRAIT].values\n",
    "Z = merged_data[condition].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Select the appropriate regression model depending on whether the dataset shows batch effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.403263628Z",
     "start_time": "2024-01-02T01:03:36.401336170Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "has_batch_effect = detect_batch_effect(X)\n",
    "has_batch_effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:36.475053437Z",
     "start_time": "2024-01-02T01:03:36.444397231Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "if has_batch_effect:\n",
    "    model_constructor1 = VariableSelection\n",
    "    model_params1 = {'modified': True, 'lamda': 3e-4}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}\n",
    "else:\n",
    "    model_constructor1 = Lasso\n",
    "    model_params1 = {'alpha': 1.0, 'random_state': 42}\n",
    "    model_constructor2 = VariableSelection\n",
    "    model_params2 = {'modified': False}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:50.733335364Z",
     "start_time": "2024-01-02T01:03:36.444563711Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "alpha for Lasso: 0.0003\n",
      "The cross-validation accuracy is 56.67% ± 20.68%\n"
     ]
    }
   ],
   "source": [
    "trait_type = 'binary'  # Remember to set this properly, either 'binary' or 'continuous'\n",
    "cv_mean1, cv_std1 = cross_validation(X, Y, Z, model_constructor1, model_params1, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The cross-validation accuracy is 95.00% ± 6.67%\n"
     ]
    }
   ],
   "source": [
    "cv_mean2, cv_std2 = cross_validation(X, Y, Z, model_constructor2, model_params2, target_type=trait_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:53.587340410Z",
     "start_time": "2024-01-02T01:03:50.732779025Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha for Lasso: 0.0003\n"
     ]
    }
   ],
   "source": [
    "normalized_X, _ = normalize_data(X)\n",
    "normalized_Z, _ = normalize_data(Z)\n",
    "\n",
    "# Train regression model on the whole dataset to identify significant genes\n",
    "model1 = ResidualizationRegressor(model_constructor1, model_params1)\n",
    "model1.fit(normalized_X, Y, normalized_Z)\n",
    "\n",
    "model2 = ResidualizationRegressor(model_constructor2, model_params2)\n",
    "model2.fit(normalized_X, Y, normalized_Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-10-14T04:50:59.072075075Z",
     "start_time": "2023-10-14T04:50:38.739499998Z"
    },
    "id": "EjJrxbvb4nlj"
   },
   "source": [
    "### 4. Discussion and report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-02T01:03:53.641570844Z",
     "start_time": "2024-01-02T01:03:53.627778553Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect of the condition on the target variable:\n",
      "Variable: Age\n",
      "Coefficient: -0.0429\n",
      "Found 50 genes with non-zero coefficients associated with the trait 'Brugada Syndrome' conditional on the factor 'Age'. These genes are identified as significant based on the regression model.\n",
      "Variable  Coefficient  Absolute Coefficient\n",
      "   SFXN5     6.968654              6.968654\n",
      "    HNMT     5.931100              5.931100\n",
      "   LIMK2    -5.185019              5.185019\n",
      "  PDLIM2     5.138317              5.138317\n",
      "   EXOC3    -4.691808              4.691808\n",
      "  KCNK15    -4.399284              4.399284\n",
      "    UGP2     3.874205              3.874205\n",
      "    LCAT     3.625004              3.625004\n",
      " RHOBTB1    -3.490019              3.490019\n",
      "   METRN     3.480335              3.480335\n",
      "   PKHD1     3.149871              3.149871\n",
      "    CHDH     3.043823              3.043823\n",
      "    KPRP    -3.011505              3.011505\n",
      " TNFSF12     2.901458              2.901458\n",
      " MADCAM1    -2.832544              2.832544\n",
      "    CES2     2.748216              2.748216\n",
      " PCDH11Y    -2.723167              2.723167\n",
      "   MKLN1     2.692490              2.692490\n",
      "    CSF2    -2.667795              2.667795\n",
      "    MYH3     2.592556              2.592556\n",
      "   APOL6     2.373610              2.373610\n",
      " FAM118A    -2.033151              2.033151\n",
      "   RNPC3    -1.876619              1.876619\n",
      " CCDC88B     1.849324              1.849324\n",
      "    ZNF7     1.746636              1.746636\n",
      "  ZNF229    -1.702903              1.702903\n",
      "    HCN4     1.618782              1.618782\n",
      "   GSTT1     1.552547              1.552547\n",
      " ALDH5A1     1.484526              1.484526\n",
      "   NXPH3    -1.078968              1.078968\n",
      "   ATP7A    -1.070324              1.070324\n",
      "   CLDN6     1.052002              1.052002\n",
      "   ACTN1     1.022854              1.022854\n",
      "   GOSR1     0.994648              0.994648\n",
      "  MRPL44     0.865972              0.865972\n",
      "     HDC    -0.851602              0.851602\n",
      "CLEC4GP1    -0.846276              0.846276\n",
      "    NET1     0.780215              0.780215\n",
      "  ZCCHC2     0.690786              0.690786\n",
      "    USP6     0.640616              0.640616\n",
      "    JPH3    -0.560973              0.560973\n",
      "    UTF1     0.548955              0.548955\n",
      "    AQP5     0.541504              0.541504\n",
      "   GRHL2     0.404603              0.404603\n",
      "    GLG1     0.377791              0.377791\n",
      "     CD2     0.360389              0.360389\n",
      "     WT1    -0.245814              0.245814\n",
      "   ERP27    -0.058313              0.058313\n",
      "    GGA2     0.026740              0.026740\n",
      "   ANXA6     0.022428              0.022428\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yxcheng/AI4Science_gold_standard/stage1/Yuxuan/../utils.py:469: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n"
     ]
    }
   ],
   "source": [
    "\n",
    "feature_cols = merged_data.columns.tolist()\n",
    "feature_cols.remove(TRAIT)\n",
    "\n",
    "threshold = 0.05\n",
    "interpret_result(model1, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Effect of the condition on the target variable:\n",
      "Variable: Age\n",
      "Coefficient: -0.0429\n",
      "Found 247 significant genes associated with the trait 'Brugada Syndrome' conditional on the factor 'Age', with corrected p-value < 0.05:\n",
      "Variable  Coefficient      p_value  corrected_p_value\n",
      "     BSN     0.330125 5.704162e-13       4.845115e-09\n",
      "    DPP6     0.340749 4.314206e-13       4.845115e-09\n",
      "     MT3     0.279684 1.196420e-12       6.774928e-09\n",
      " PPP2R2C     0.348435 1.795095e-11       7.623769e-08\n",
      "   GAP43     0.341790 5.879802e-11       1.664768e-07\n",
      "    SCG3     0.288491 5.537623e-11       1.664768e-07\n",
      "   KIF5A     0.294121 7.789659e-11       1.890439e-07\n",
      "  SNAP25     0.273705 1.313348e-10       2.788894e-07\n",
      "    MAP2     0.305473 3.887751e-10       7.338346e-07\n",
      "   GNAO1     0.277382 5.207726e-10       8.846884e-07\n",
      "    NNAT     0.310292 5.988944e-10       9.249108e-07\n",
      "    CHD5     0.285377 1.723972e-09       2.440569e-06\n",
      "    RTN1     0.316324 2.505213e-09       2.862876e-06\n",
      "   TRIM9     0.250064 2.527851e-09       2.862876e-06\n",
      "   PDZD4     0.273653 2.319458e-09       2.862876e-06\n",
      "    HPCA     0.292228 3.948644e-09       4.192472e-06\n",
      "    DNM1     0.251860 5.416192e-09       5.412369e-06\n",
      "     SYP     0.250300 8.322826e-09       7.854898e-06\n",
      "   NRXN2     0.304611 9.009515e-09       8.055454e-06\n",
      "  TCEAL2     0.273106 2.015561e-08       1.712017e-05\n",
      "    DLG2     0.299978 3.216514e-08       2.483734e-05\n",
      "  SCAMP5     0.249398 3.208037e-08       2.483734e-05\n",
      "   GPM6B     0.219235 3.786019e-08       2.796387e-05\n",
      "  SYNGR1     0.268259 5.345124e-08       3.783457e-05\n",
      "    SNCB     0.240309 8.696716e-08       5.909592e-05\n",
      "    NEFL     0.268417 9.724533e-08       6.353860e-05\n",
      " PPP2R2B     0.273955 1.041832e-07       6.452689e-05\n",
      "   CSPG5     0.259623 1.063547e-07       6.452689e-05\n",
      "    SCG2     0.249428 1.405840e-07       8.235316e-05\n",
      "KIAA0513     0.222348 1.610977e-07       9.122428e-05\n",
      "   CAMKV     0.229741 4.958029e-07       2.506581e-04\n",
      "  MAPK10     0.228232 4.980424e-07       2.506581e-04\n",
      "   GRIN1     0.250916 4.749839e-07       2.506581e-04\n",
      "   SYT11     0.243537 5.016702e-07       2.506581e-04\n",
      "   ABCG4     0.213292 5.294595e-07       2.569845e-04\n",
      "   GRIK5     0.213905 5.884144e-07       2.776662e-04\n",
      "    SCG5     0.221402 7.195197e-07       3.303568e-04\n",
      "    GRM3     0.261928 8.671949e-07       3.852880e-04\n",
      "    MATK     0.204453 8.845204e-07       3.852880e-04\n",
      " CAMK2N1     0.175927 9.906508e-07       4.207294e-04\n",
      "     CPE     0.205378 1.021386e-06       4.232026e-04\n",
      "ATP6V0E2     0.174795 1.253332e-06       5.069429e-04\n",
      "    ATL1     0.221436 1.663875e-06       6.573467e-04\n",
      "     NTM     0.239310 1.838437e-06       7.098038e-04\n",
      " HSPA12A     0.190252 1.934029e-06       7.142452e-04\n",
      "   CRMP1     0.258343 1.905958e-06       7.142452e-04\n",
      " PLEKHB1     0.115160 2.183464e-06       7.892059e-04\n",
      "  CLSTN1     0.220236 2.508671e-06       8.878606e-04\n",
      "     IDS     0.194816 2.627980e-06       9.111046e-04\n",
      "  FRMPD4     0.203314 2.929093e-06       9.951888e-04\n",
      "  SH3GL2     0.195848 3.241120e-06       1.074190e-03\n",
      "    CA11     0.144604 3.348980e-06       1.074190e-03\n",
      "    APC2     0.245338 3.351311e-06       1.074190e-03\n",
      "  CELSR2     0.217328 4.046098e-06       1.272873e-03\n",
      "   TUBB3     0.212586 4.508757e-06       1.343768e-03\n",
      " KBTBD11     0.210426 4.418979e-06       1.343768e-03\n",
      "  MGAT5B     0.191377 4.450409e-06       1.343768e-03\n",
      "   WDR17     0.180019 5.384147e-06       1.576998e-03\n",
      "    SYN1     0.200736 5.847763e-06       1.683759e-03\n",
      "  PCSK1N     0.218409 6.148395e-06       1.697479e-03\n",
      "   BAALC     0.163073 6.195178e-06       1.697479e-03\n",
      "  SH3GL3     0.214328 6.039072e-06       1.697479e-03\n",
      "   VAMP2     0.209190 6.352575e-06       1.701269e-03\n",
      "  STXBP1     0.223867 6.509447e-06       1.701269e-03\n",
      " CCDC136     0.205548 6.444706e-06       1.701269e-03\n",
      "   CLIP3     0.206951 6.991501e-06       1.799570e-03\n",
      "  RAB40B     0.135350 8.504208e-06       2.127007e-03\n",
      "   KNDC1     0.132176 8.514036e-06       2.127007e-03\n",
      "   GABRD     0.137986 9.009605e-06       2.206102e-03\n",
      "    LRP8     0.173825 9.090364e-06       2.206102e-03\n",
      "   ITM2C     0.151201 9.326483e-06       2.231525e-03\n",
      "   FABP6     0.163878 1.089933e-05       2.571636e-03\n",
      "   DOC2A     0.200117 1.142358e-05       2.658408e-03\n",
      "   LSAMP     0.212827 1.214657e-05       2.788459e-03\n",
      "  NDFIP1     0.194657 1.257484e-05       2.848285e-03\n",
      "   PAQR6     0.125222 1.388907e-05       3.024968e-03\n",
      "   OLFM2     0.196563 1.363372e-05       3.024968e-03\n",
      "    ENHO     0.152964 1.375422e-05       3.024968e-03\n",
      "    SNCG     0.113796 1.436721e-05       3.089495e-03\n",
      "    SV2B     0.169016 1.586794e-05       3.369557e-03\n",
      "  NAP1L5     0.195197 1.618710e-05       3.394894e-03\n",
      "   PTPRA     0.197173 1.778093e-05       3.683687e-03\n",
      "   KIFC2     0.166818 1.810715e-05       3.706076e-03\n",
      "   MAP1A     0.155509 1.860531e-05       3.762702e-03\n",
      "     APC     0.169761 1.889137e-05       3.775606e-03\n",
      "    NRGN     0.185256 1.935318e-05       3.822928e-03\n",
      "  CACNB3     0.164978 2.156075e-05       4.194513e-03\n",
      "    SHC3     0.119470 2.172811e-05       4.194513e-03\n",
      "   STMN2     0.198133 2.223742e-05       4.244599e-03\n",
      "    NRG3     0.181395 2.340529e-05       4.417879e-03\n",
      "  VSTM2A     0.170399 2.568539e-05       4.733113e-03\n",
      "    KLC1     0.226301 2.590060e-05       4.733113e-03\n",
      "   SYNJ1     0.173186 2.591120e-05       4.733113e-03\n",
      "   AP3B2     0.182733 2.846517e-05       5.144322e-03\n",
      "SLC25A22     0.148583 2.972414e-05       5.315302e-03\n",
      "   PLCD4     0.112900 3.097765e-05       5.481753e-03\n",
      "  DBNDD1     0.148482 3.199324e-05       5.489911e-03\n",
      "    DTNB     0.154810 3.173785e-05       5.489911e-03\n",
      "    PALM     0.174259 3.149095e-05       5.489911e-03\n",
      "   CEND1     0.144296 3.298909e-05       5.604187e-03\n",
      "    SYT1     0.214982 3.568059e-05       6.001405e-03\n",
      "  KATNB1     0.167362 3.867642e-05       6.441520e-03\n",
      " CNTNAP2     0.163160 3.957464e-05       6.527126e-03\n",
      "   NTRK2     0.152313 4.128876e-05       6.604206e-03\n",
      "    GNG3     0.186094 4.198577e-05       6.604206e-03\n",
      "  GABRB1     0.176060 4.169033e-05       6.604206e-03\n",
      "   EXTL1     0.122895 4.152849e-05       6.604206e-03\n",
      "  CKMT1B     0.174506 4.111156e-05       6.604206e-03\n",
      "   MCTP1     0.162047 4.358370e-05       6.649376e-03\n",
      "   CPLX1     0.123740 4.383860e-05       6.649376e-03\n",
      " COL11A1     0.151173 4.276834e-05       6.649376e-03\n",
      "HSD11B1L     0.159868 4.309800e-05       6.649376e-03\n",
      "   TTYH1     0.144639 4.600251e-05       6.915846e-03\n",
      "   CHRM3     0.156486 4.862624e-05       7.246163e-03\n",
      "    DLL3     0.182818 5.077945e-05       7.501229e-03\n",
      "   SFXN5     0.114674 5.134219e-05       7.518976e-03\n",
      "  FAM98C     0.129173 5.342667e-05       7.757370e-03\n",
      "   TTBK1     0.159224 5.577080e-05       8.029105e-03\n",
      "  GABRA3     0.172105 5.764402e-05       8.229048e-03\n",
      "  LRRC3B     0.144124 5.830184e-05       8.253597e-03\n",
      "  ABLIM2     0.123431 6.014711e-05       8.444455e-03\n",
      "   NDRG3     0.187842 6.674602e-05       9.251407e-03\n",
      "    PMP2     0.099771 6.824234e-05       9.251407e-03\n",
      "     INA     0.175892 6.916227e-05       9.251407e-03\n",
      "  CAMK2A     0.131720 6.814689e-05       9.251407e-03\n",
      "   RASD2     0.143697 6.864237e-05       9.251407e-03\n",
      "   STOX2     0.159730 6.733327e-05       9.251407e-03\n",
      "  PPFIA2     0.167404 7.012768e-05       9.307258e-03\n",
      " ANKRD35     0.114046 7.302233e-05       9.616306e-03\n",
      "   DDHD2     0.137829 7.921766e-05       1.035192e-02\n",
      "  TAGLN3     0.163728 8.010262e-05       1.038766e-02\n",
      "  IGSF21     0.187154 8.436511e-05       1.085753e-02\n",
      "    MAPT     0.206703 8.894255e-05       1.136057e-02\n",
      " SLC6A17     0.174970 9.116135e-05       1.155708e-02\n",
      "     TNR     0.126782 9.202809e-05       1.158054e-02\n",
      "  PARD6A     0.136769 9.718753e-05       1.213987e-02\n",
      "    ULK2     0.187225 9.954189e-05       1.220017e-02\n",
      "  ZNF706     0.159178 9.982482e-05       1.220017e-02\n",
      "  CACNG2     0.176683 9.963065e-05       1.220017e-02\n",
      "   AGAP3     0.163295 1.015935e-04       1.232764e-02\n",
      "    FSD1     0.196282 1.050931e-04       1.266186e-02\n",
      "    MXD4     0.136422 1.172472e-04       1.402673e-02\n",
      " PPP2R5B     0.174135 1.258176e-04       1.474488e-02\n",
      "    GAD2     0.109645 1.243270e-04       1.474488e-02\n",
      "   OLFM1     0.130172 1.258540e-04       1.474488e-02\n",
      "   NR2F1     0.148320 1.429272e-04       1.663046e-02\n",
      "  TMEFF2     0.165659 1.495104e-04       1.727811e-02\n",
      "   KCNC2     0.137266 1.522853e-04       1.747989e-02\n",
      "   CPEB1     0.142127 1.538985e-04       1.754649e-02\n",
      "   PRMT8     0.150023 1.603464e-04       1.815976e-02\n",
      "  HPCAL4     0.151855 1.653759e-04       1.860534e-02\n",
      "  KCNMA1     0.131146 1.679526e-04       1.864822e-02\n",
      "   KIF1B     0.182362 1.679243e-04       1.864822e-02\n",
      "   BEND6     0.148147 1.759053e-04       1.932066e-02\n",
      "   KCNH3     0.128195 1.762834e-04       1.932066e-02\n",
      " CCDC85A     0.143442 1.776603e-04       1.934675e-02\n",
      " TBC1D9B     0.113830 1.802853e-04       1.950756e-02\n",
      "  ELOVL7     0.080796 1.844814e-04       1.980278e-02\n",
      " ST3GAL5     0.145637 1.853451e-04       1.980278e-02\n",
      "  GALNT9     0.099611 1.881947e-04       1.998157e-02\n",
      "  GABRA2     0.145766 1.908313e-04       2.013567e-02\n",
      "   CLCN4     0.132858 1.926959e-04       2.015581e-02\n",
      "   BRSK1     0.134134 1.933952e-04       2.015581e-02\n",
      "   DGCR5     0.110990 1.994284e-04       2.065786e-02\n",
      " SNRNP25     0.090461 2.077730e-04       2.135815e-02\n",
      "    SYN2     0.125524 2.087034e-04       2.135815e-02\n",
      " FAM171B     0.141650 2.114034e-04       2.147810e-02\n",
      "   MAPK4     0.100590 2.124041e-04       2.147810e-02\n",
      "   UCHL1     0.195550 2.208904e-04       2.207345e-02\n",
      "   TIAM1     0.123126 2.197467e-04       2.207345e-02\n",
      "  SERGEF     0.101009 2.247409e-04       2.232689e-02\n",
      " ATP6V1H     0.162742 2.289043e-04       2.237600e-02\n",
      "   HABP4     0.116439 2.276195e-04       2.237600e-02\n",
      "   ATP9A     0.163911 2.291867e-04       2.237600e-02\n",
      "    BCAN     0.105311 2.364264e-04       2.272715e-02\n",
      "  EMX2OS     0.091902 2.367968e-04       2.272715e-02\n",
      " OSBPL1A     0.119973 2.355505e-04       2.272715e-02\n",
      "   KIF3C     0.155821 2.505187e-04       2.390905e-02\n",
      "    GIT1     0.165683 2.554552e-04       2.424398e-02\n",
      "  TSPAN7     0.155981 2.630081e-04       2.482212e-02\n",
      "    CHL1     0.171580 2.709342e-04       2.531167e-02\n",
      "   PTPRN     0.154096 2.711752e-04       2.531167e-02\n",
      "    PKP4     0.131940 2.773839e-04       2.574971e-02\n",
      "   NDST3     0.123883 2.818296e-04       2.602022e-02\n",
      "   PDE8B     0.121992 2.883271e-04       2.647622e-02\n",
      "   PRKCB     0.136047 2.919714e-04       2.666673e-02\n",
      "   RUFY3     0.179072 2.955578e-04       2.684992e-02\n",
      "TMEM170B     0.137052 2.998065e-04       2.709102e-02\n",
      "  MAP6D1     0.131288 3.041205e-04       2.719158e-02\n",
      "   KCNT1     0.101571 3.027531e-04       2.719158e-02\n",
      "  ELAVL3     0.162482 3.118247e-04       2.773444e-02\n",
      "    CUX2     0.143913 3.147176e-04       2.784595e-02\n",
      "  SCUBE2     0.085749 3.246365e-04       2.857474e-02\n",
      "SERPINI1     0.122117 3.309389e-04       2.897933e-02\n",
      "    NGRN     0.131240 3.396298e-04       2.958785e-02\n",
      "  RGS7BP     0.108910 3.432680e-04       2.964643e-02\n",
      "   KCNB1     0.126034 3.437925e-04       2.964643e-02\n",
      "  ELOVL4     0.163925 3.599559e-04       3.088349e-02\n",
      "  SLC9A2     0.101549 3.648591e-04       3.114687e-02\n",
      "  KCNAB1     0.115185 3.714489e-04       3.155087e-02\n",
      "  SPTBN2     0.123930 3.790768e-04       3.203859e-02\n",
      "   RIMS3     0.123147 3.814291e-04       3.207781e-02\n",
      "     NSF     0.147245 3.980840e-04       3.331355e-02\n",
      "   KCNK2     0.105354 4.184783e-04       3.484858e-02\n",
      "   SPEF1     0.117456 4.236586e-04       3.495083e-02\n",
      "   ZBTB4     0.124233 4.238210e-04       3.495083e-02\n",
      " FAM184A     0.108342 4.346211e-04       3.566833e-02\n",
      " PHACTR3     0.151142 4.383866e-04       3.580438e-02\n",
      "  GPR162     0.137856 4.418980e-04       3.591848e-02\n",
      "    LIX1     0.110901 4.455228e-04       3.604067e-02\n",
      "    PLD6     0.092033 4.567233e-04       3.677164e-02\n",
      "   DACT3     0.147719 4.595173e-04       3.682207e-02\n",
      "    WDR7     0.146797 4.624502e-04       3.688312e-02\n",
      "   MATR3     0.145534 4.766346e-04       3.783677e-02\n",
      "   KCNF1     0.097256 4.843072e-04       3.826702e-02\n",
      "  PKNOX2     0.137160 4.878156e-04       3.836579e-02\n",
      "  FBXW11     0.149783 4.913618e-04       3.842838e-02\n",
      "  PTPN13     0.129202 4.931356e-04       3.842838e-02\n",
      "  MAPRE3     0.130784 5.052709e-04       3.918239e-02\n",
      "   KCNJ3     0.116258 5.074244e-04       3.918239e-02\n",
      "  CNRIP1     0.167718 5.161968e-04       3.950068e-02\n",
      "  ATP1A2     0.078218 5.138885e-04       3.950068e-02\n",
      "   KCTD2     0.151808 5.274591e-04       4.018150e-02\n",
      "  SLC8A2     0.164603 5.330220e-04       4.042401e-02\n",
      "   FBXW7     0.134916 5.387017e-04       4.067317e-02\n",
      "   CLCN6     0.120421 5.438119e-04       4.080584e-02\n",
      " PHYHIPL     0.163147 5.452629e-04       4.080584e-02\n",
      "   DUSP8     0.125627 5.553908e-04       4.138148e-02\n",
      "    FRZB     0.089496 5.681140e-04       4.197745e-02\n",
      "    HCN1     0.113047 5.683314e-04       4.197745e-02\n",
      "   GRIK1     0.124642 5.750766e-04       4.229178e-02\n",
      "   SULF2     0.077999 5.777708e-04       4.230677e-02\n",
      "    BEX2     0.141698 5.856830e-04       4.241106e-02\n",
      "  TCEAL3     0.119400 5.866846e-04       4.241106e-02\n",
      "   PLCD3     0.115103 5.820810e-04       4.241106e-02\n",
      "    GAD1     0.129978 6.108239e-04       4.396897e-02\n",
      "  CAMTA1     0.133257 6.208426e-04       4.450158e-02\n",
      "  LANCL2     0.134767 6.346478e-04       4.522644e-02\n",
      "  B3GAT1     0.126546 6.362797e-04       4.522644e-02\n",
      "  ANKS1B     0.159587 6.502776e-04       4.583783e-02\n",
      "  PPP4R4     0.096205 6.490270e-04       4.583783e-02\n",
      "    EPN2     0.136880 6.613098e-04       4.642286e-02\n",
      "  SHANK2     0.148388 6.810770e-04       4.751977e-02\n",
      "   APBA2     0.148777 6.825302e-04       4.751977e-02\n",
      " PPP1R9B     0.107132 7.001796e-04       4.854960e-02\n",
      "  ZNF540     0.106253 7.079638e-04       4.888979e-02\n",
      "   PHF14     0.145137 7.133620e-04       4.906313e-02\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yxcheng/AI4Science_gold_standard/stage1/Yuxuan/../utils.py:478: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"
     ]
    }
   ],
   "source": [
    "\n",
    "interpret_result(model2, feature_cols, TRAIT, condition, threshold=threshold, save_output=True,\n",
    "                 output_dir=OUTPUT_DIR, model_id=2)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
