{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Gold standard curation: Preprocessing and single-step regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n",
    "\n",
    "1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n",
    "2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import io\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import mygene\n",
    "import re\n",
    "import tempfile\n",
    "import shutil\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import json\n",
    "from typing import Callable, Optional, List, Tuple, Union, Any\n",
    "from sklearn.linear_model import Lasso, LogisticRegression\n",
    "from sparse_lmm import VariableSelection\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from sklearn.metrics import accuracy_score, mean_squared_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant_filepaths(cohort_dir):\n",
    "    \"\"\"Find the file paths of a SOFT file and a matrix file from the given data directory of a cohort.\n",
    "    If there are multiple SOFT files or matrix files, simply choose the first one. May be replaced by better\n",
    "    strategies later.\n",
    "    \"\"\"\n",
    "    files = os.listdir(cohort_dir)\n",
    "    soft_files = [f for f in files if 'soft' in f.lower()]\n",
    "    matrix_files = [f for f in files if 'matrix' in f.lower()]\n",
    "    assert len(soft_files) > 0 and len(matrix_files) > 0\n",
    "    soft_file_path = os.path.join(cohort_dir, soft_files[0])\n",
    "    matrix_file_path = os.path.join(cohort_dir, matrix_files[0])\n",
    "\n",
    "    return soft_file_path, matrix_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n",
    "\n",
    "\n",
    "def get_background_and_clinical_data(file_path,\n",
    "                                     prefixes_a=['!Series_title', '!Series_summary', '!Series_overall_design'],\n",
    "                                     prefixes_b=['!Sample_geo_accession', '!Sample_characteristics_ch1']):\n",
    "    \"\"\"Extract from a matrix file the background information about the dataset, and sample characteristics data\"\"\"\n",
    "    background_info, clinical_data = filter_content_by_prefix(file_path, prefixes_a, prefixes_b, unselect=False,\n",
    "                                                              source_type='file',\n",
    "                                                              return_df_a=False, return_df_b=True)\n",
    "    return background_info, clinical_data\n",
    "\n",
    "\n",
    "def get_gene_annotation(file_path, prefixes=['^', '!', '#']):\n",
    "    \"\"\"Extract from a SOFT file the gene annotation data\"\"\"\n",
    "    gene_metadata = filter_content_by_prefix(file_path, prefixes_a=prefixes, unselect=True, source_type='file',\n",
    "                                             return_df_a=True)\n",
    "    return gene_metadata[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def line_generator(source, source_type):\n",
    "    \"\"\"Generator that yields lines from a file or a string.\n",
    "\n",
    "    Parameters:\n",
    "    - source: File path or string content.\n",
    "    - source_type: 'file' or 'string'.\n",
    "    \"\"\"\n",
    "    if source_type == 'file':\n",
    "        with gzip.open(source, 'rt') as f:\n",
    "            for line in f:\n",
    "                yield line.strip()\n",
    "    elif source_type == 'string':\n",
    "        for line in source.split('\\n'):\n",
    "            yield line.strip()\n",
    "    else:\n",
    "        raise ValueError(\"source_type must be 'file' or 'string'\")\n",
    "\n",
    "\n",
    "def filter_content_by_prefix(\n",
    "    source: str,\n",
    "    prefixes_a: List[str],\n",
    "    prefixes_b: Optional[List[str]] = None,\n",
    "    unselect: bool = False,\n",
    "    source_type: str = 'file',\n",
    "    return_df_a: bool = True,\n",
    "    return_df_b: bool = True\n",
    ") -> Tuple[Union[str, pd.DataFrame], Optional[Union[str, pd.DataFrame]]]:\n",
    "    \"\"\"\n",
    "    Filters rows from a file or a list of strings based on specified prefixes.\n",
    "\n",
    "    Parameters:\n",
    "    - source (str): File path or string content to filter.\n",
    "    - prefixes_a (List[str]): Primary list of prefixes to filter by.\n",
    "    - prefixes_b (Optional[List[str]]): Optional secondary list of prefixes to filter by.\n",
    "    - unselect (bool): If True, selects rows that do not start with the specified prefixes.\n",
    "    - source_type (str): 'file' if source is a file path, 'string' if source is a string of text.\n",
    "    - return_df_a (bool): If True, returns filtered content for prefixes_a as a pandas DataFrame.\n",
    "    - return_df_b (bool): If True, and if prefixes_b is provided, returns filtered content for prefixes_b as a pandas DataFrame.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple: A tuple where the first element is the filtered content for prefixes_a, and the second element is the filtered content for prefixes_b.\n",
    "    \"\"\"\n",
    "    filtered_lines_a = []\n",
    "    filtered_lines_b = []\n",
    "    prefix_set_a = set(prefixes_a)\n",
    "    if prefixes_b is not None:\n",
    "        prefix_set_b = set(prefixes_b)\n",
    "\n",
    "    # Use generator to get lines\n",
    "    for line in line_generator(source, source_type):\n",
    "        matched_a = any(line.startswith(prefix) for prefix in prefix_set_a)\n",
    "        if matched_a != unselect:\n",
    "            filtered_lines_a.append(line)\n",
    "        if prefixes_b is not None:\n",
    "            matched_b = any(line.startswith(prefix) for prefix in prefix_set_b)\n",
    "            if matched_b != unselect:\n",
    "                filtered_lines_b.append(line)\n",
    "\n",
    "    filtered_content_a = '\\n'.join(filtered_lines_a)\n",
    "    if return_df_a:\n",
    "        filtered_content_a = pd.read_csv(io.StringIO(filtered_content_a), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "    filtered_content_b = None\n",
    "    if filtered_lines_b:\n",
    "        filtered_content_b = '\\n'.join(filtered_lines_b)\n",
    "        if return_df_b:\n",
    "            filtered_content_b = pd.read_csv(io.StringIO(filtered_content_b), delimiter='\\t', low_memory=False, on_bad_lines='skip')\n",
    "\n",
    "    return filtered_content_a, filtered_content_b\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_values_by_row(dataframe, max_len=30):\n",
    "    \"\"\"\n",
    "    Organize the unique values in each row of the given dataframe, to get a dictionary\n",
    "    :param dataframe:\n",
    "    :param max_len:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    if '!Sample_geo_accession' in dataframe.columns:\n",
    "        dataframe = dataframe.drop(columns=['!Sample_geo_accession'])\n",
    "    unique_values_dict = {}\n",
    "    for index, row in dataframe.iterrows():\n",
    "        unique_values = list(row.unique())[:max_len]\n",
    "        unique_values_dict[index] = unique_values\n",
    "    return unique_values_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_select_clinical_features(clinical_df: pd.DataFrame, trait: str, trait_row: int,\n",
    "                                 convert_trait: Callable,\n",
    "                                 age_row: Optional[int] = None,\n",
    "                                 convert_age: Optional[Callable] = None,\n",
    "                                 gender_row: Optional[int] = None,\n",
    "                                 convert_gender: Optional[Callable] = None) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Extracts and processes specific clinical features from a DataFrame representing\n",
    "    sample characteristics in the GEO database series.\n",
    "\n",
    "    Parameters:\n",
    "    - clinical_df (pd.DataFrame): DataFrame containing clinical data.\n",
    "    - trait (str): The trait of interest.\n",
    "    - trait_row (int): Row identifier for the trait in the DataFrame.\n",
    "    - convert_trait (Callable): Function to convert trait data into a desired format.\n",
    "    - age_row (int, optional): Row identifier for age data. Default is None.\n",
    "    - convert_age (Callable, optional): Function to convert age data. Default is None.\n",
    "    - gender_row (int, optional): Row identifier for gender data. Default is None.\n",
    "    - convert_gender (Callable, optional): Function to convert gender data. Default is None.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the selected and processed clinical features.\n",
    "    \"\"\"\n",
    "    feature_list = []\n",
    "\n",
    "    trait_data = get_feature_data(clinical_df, trait_row, trait, convert_trait)\n",
    "    feature_list.append(trait_data)\n",
    "    if age_row is not None:\n",
    "        age_data = get_feature_data(clinical_df, age_row, 'Age', convert_age)\n",
    "        feature_list.append(age_data)\n",
    "    if gender_row is not None:\n",
    "        gender_data = get_feature_data(clinical_df, gender_row, 'Gender', convert_gender)\n",
    "        feature_list.append(gender_data)\n",
    "\n",
    "    selected_clinical_df = pd.concat(feature_list, axis=0)\n",
    "    return selected_clinical_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_data(clinical_df, row_id, feature, convert_fn):\n",
    "    \"\"\"select the row corresponding to a feature in the sample characteristics dataframe, and convert the feature into\n",
    "    a binary or continuous variable\"\"\"\n",
    "    clinical_df = clinical_df.iloc[row_id:row_id + 1].drop(columns=['!Sample_geo_accession'], errors='ignore')\n",
    "    clinical_df.index = [feature]\n",
    "    clinical_df = clinical_df.applymap(convert_fn)\n",
    "\n",
    "    return clinical_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_cohort_info(cohort: str, info_path: str, is_available: bool, is_biased: Optional[bool] = None,\n",
    "                     df: Optional[pd.DataFrame] = None, note: str = '') -> None:\n",
    "    \"\"\"\n",
    "    Add or update information about the usability and quality of a dataset for statistical analysis.\n",
    "\n",
    "    Parameters:\n",
    "    cohort (str): A unique identifier for the dataset.\n",
    "    info_path (str): File path to the JSON file where records are stored.\n",
    "    is_available (bool): Indicates whether both the genetic data and trait data are available in the dataset, and can be\n",
    "     preprocessed into a dataframe.\n",
    "    is_biased (bool, optional): Indicates whether the dataset is too biased to be usable.\n",
    "        Required if `is_available` is True.\n",
    "    df (pandas.DataFrame, optional): The preprocessed dataset. Required if `is_available` is True.\n",
    "    note (str, optional): Additional notes about the dataset.\n",
    "\n",
    "    Returns:\n",
    "    None: The function does not return a value but updates or creates a record in the specified JSON file.\n",
    "    \"\"\"\n",
    "    if is_available:\n",
    "        assert (df is not None) and (is_biased is not None), \"'df' and 'is_biased' should be provided if this cohort \" \\\n",
    "                                                             \"is relevant.\"\n",
    "    is_usable = is_available and (not is_biased)\n",
    "    new_record = {\"is_usable\": is_usable,\n",
    "                  \"is_available\": is_available,\n",
    "                  \"is_biased\": is_biased if is_available else None,\n",
    "                  \"has_age\": \"Age\" in df.columns if is_available else None,\n",
    "                  \"has_gender\": \"Gender\" in df.columns if is_available else None,\n",
    "                  \"sample_size\": len(df) if is_available else None,\n",
    "                  \"note\": note}\n",
    "\n",
    "    if not os.path.exists(info_path):\n",
    "        with open(info_path, 'w') as file:\n",
    "            json.dump({}, file)\n",
    "        print(f\"A new JSON file was created at: {info_path}\")\n",
    "\n",
    "    with open(info_path, \"r\") as file:\n",
    "        records = json.load(file)\n",
    "    records[cohort] = new_record\n",
    "\n",
    "    temp_path = info_path + \".tmp\"\n",
    "    try:\n",
    "        with open(temp_path, 'w') as file:\n",
    "            json.dump(records, file)\n",
    "        os.replace(temp_path, info_path)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        if os.path.exists(temp_path):\n",
    "            os.remove(temp_path)\n",
    "        raise\n",
    "\n",
    "\n",
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A pandas DataFrame with the JSON data.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genetic_data(file_path):\n",
    "    \"\"\"Read the gene expression data into a dataframe, and adjust its format\"\"\"\n",
    "    genetic_data = pd.read_csv(file_path, compression='gzip', skiprows=52, comment='!', delimiter='\\t')\n",
    "    genetic_data = genetic_data.dropna()\n",
    "    genetic_data = genetic_data.rename(columns={'ID_REF': 'ID'}).astype({'ID': 'str'})\n",
    "    genetic_data.set_index('ID', inplace=True)\n",
    "\n",
    "    return genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preview_df(df, n=5):\n",
    "    return df.head(n).to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols(gene_symbols, batch_size=1000):\n",
    "    \"\"\"Normalize human gene symbols in batches using the 'mygenes' library\"\"\"\n",
    "    mg = mygene.MyGeneInfo()\n",
    "    normalized_genes = {}\n",
    "\n",
    "    # Process in batches\n",
    "    for i in range(0, len(gene_symbols), batch_size):\n",
    "        batch = gene_symbols[i:i + batch_size]\n",
    "        results = mg.querymany(batch, scopes='symbol', fields='symbol', species='human')\n",
    "\n",
    "        # Update the normalized_genes dictionary with results from this batch\n",
    "        for gene in results:\n",
    "            normalized_genes[gene['query']] = gene.get('symbol', None)\n",
    "\n",
    "    # Return the normalized symbols in the same order as the input\n",
    "    return [normalized_genes.get(symbol) for symbol in gene_symbols]\n",
    "\n",
    "\n",
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_gene_symbols_in_index(gene_df):\n",
    "    \"\"\"Normalize the human gene symbols at the index of a dataframe, and replace the index with its normalized version.\n",
    "    Remove the rows where the index failed to be normalized.\"\"\"\n",
    "    normalized_gene_list = normalize_gene_symbols(gene_df.index.tolist())\n",
    "    assert len(normalized_gene_list) == len(gene_df.index)\n",
    "    gene_df.index = normalized_gene_list\n",
    "    gene_df = gene_df[gene_df.index.notnull()]\n",
    "    return gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_mapping(annotation, prob_col, gene_col):\n",
    "    \"\"\"Process the gene annotation to get the mapping between gene names and gene probes.\n",
    "    \"\"\"\n",
    "    mapping_data = annotation.loc[:, [prob_col, gene_col]]\n",
    "    mapping_data = mapping_data.dropna()\n",
    "    mapping_data = mapping_data.rename(columns={gene_col: 'Gene'}).astype({'ID': 'str'})\n",
    "\n",
    "    return mapping_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_gene_mapping(expression_df, mapping_df):\n",
    "    \"\"\"\n",
    "    Converts measured data about gene probes into gene expression data.\n",
    "    Handles the potential many-to-many relationship between probes and genes.\n",
    "\n",
    "    Parameters:\n",
    "    expression_df (DataFrame): A DataFrame with gene expression data, indexed by 'ID'.\n",
    "    mapping_df (DataFrame): A DataFrame mapping 'ID' to 'Gene', with 'ID' as a column.\n",
    "\n",
    "    Returns:\n",
    "    DataFrame: A DataFrame with mean gene expression values, indexed by 'Gene'.\n",
    "    \"\"\"\n",
    "\n",
    "    # Define a regex pattern for splitting gene names\n",
    "    split_pattern = r';|\\|+|/{2,}|,|\\[|\\]|\\(|\\)'\n",
    "\n",
    "    # Split the 'Gene' column in 'mapping_df' using the regex pattern\n",
    "    mapping_df['Gene'] = mapping_df['Gene'].str.split(split_pattern)\n",
    "    mapping_df = mapping_df.explode('Gene')\n",
    "\n",
    "    # Set 'ID' as the index of 'mapping_df' for merging\n",
    "    mapping_df.set_index('ID', inplace=True)\n",
    "\n",
    "    # Merge 'mapping_df' with 'expression_df' by their indices\n",
    "    merged_df = mapping_df.join(expression_df)\n",
    "\n",
    "    # Group by 'Gene' and calculate the mean expression values\n",
    "    gene_expression_df = merged_df.groupby('Gene').mean().dropna()\n",
    "\n",
    "    return gene_expression_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geo_merge_clinical_genetic_data(clinical_df, genetic_df):\n",
    "    \"\"\"\n",
    "    Merge the clinical features and gene expression features from two dataframes into one dataframe\n",
    "    \"\"\"\n",
    "    if 'ID' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.rename(columns={'ID': 'Gene'})\n",
    "    if 'Gene' in genetic_df.columns:\n",
    "        genetic_df = genetic_df.set_index('Gene')\n",
    "    merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n",
    "    return merged_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_binary_variable_biased(dataframe, col_name, min_proportion=0.1, min_num=5):\n",
    "    \"\"\"\n",
    "    Check if the distribution of a binary variable in the dataset is too biased to be usable for analysis\n",
    "    :param dataframe:\n",
    "    :param col_name:\n",
    "    :param min_proportion:\n",
    "    :param min_num:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    label_counter = dataframe[col_name].value_counts()\n",
    "    total_samples = len(dataframe)\n",
    "    rare_label_num = label_counter.min()\n",
    "    rare_label = label_counter.idxmin()\n",
    "    rare_label_proportion = rare_label_num / total_samples\n",
    "\n",
    "    print(\n",
    "        f\"For the feature \\'{col_name}\\', the least common label is '{rare_label}' with {rare_label_num} occurrences. This represents {rare_label_proportion:.2%} of the dataset.\")\n",
    "\n",
    "    biased = (len(label_counter) < 2) or ((rare_label_proportion < min_proportion) and (rare_label_num < min_num))\n",
    "    return bool(biased)\n",
    "\n",
    "\n",
    "def judge_continuous_variable_biased(dataframe, col_name):\n",
    "    \"\"\"Check if the distribution of a continuous variable in the dataset is too biased to be usable for analysis.\n",
    "    As a starting point, we consider it biased if all values are the same. For the next step, maybe ask GPT to judge\n",
    "    based on quartile statistics combined with its common sense knowledge about this feature.\n",
    "    \"\"\"\n",
    "    quartiles = dataframe[col_name].quantile([0.25, 0.5, 0.75])\n",
    "    min_value = dataframe[col_name].min()\n",
    "    max_value = dataframe[col_name].max()\n",
    "\n",
    "    # Printing quartile information\n",
    "    print(f\"Quartiles for '{col_name}':\")\n",
    "    print(f\"  25%: {quartiles[0.25]}\")\n",
    "    print(f\"  50% (Median): {quartiles[0.5]}\")\n",
    "    print(f\"  75%: {quartiles[0.75]}\")\n",
    "    print(f\"Min: {min_value}\")\n",
    "    print(f\"Max: {max_value}\")\n",
    "\n",
    "    biased = min_value == max_value\n",
    "\n",
    "    return bool(biased)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "def judge_and_remove_biased_features(df, trait, trait_type):\n",
    "    assert trait_type in [\"binary\", \"continuous\"], f\"The trait must be either a binary or a continuous variable!\"\n",
    "    if trait_type == \"binary\":\n",
    "        trait_biased = judge_binary_variable_biased(df, trait)\n",
    "    else:\n",
    "        trait_biased = judge_continuous_variable_biased(df, trait)\n",
    "    if trait_biased:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is severely biased.\\n\")\n",
    "    else:\n",
    "        print(f\"The distribution of the feature \\'{trait}\\' in this dataset is fine.\\n\")\n",
    "    if \"Age\" in df.columns:\n",
    "        age_biased = judge_continuous_variable_biased(df, 'Age')\n",
    "        if age_biased:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Age')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Age\\' in this dataset is fine.\\n\")\n",
    "    if \"Gender\" in df.columns:\n",
    "        gender_biased = judge_binary_variable_biased(df, 'Gender')\n",
    "        if gender_biased:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is severely biased.\\n\")\n",
    "            df = df.drop(columns='Gender')\n",
    "        else:\n",
    "            print(f\"The distribution of the feature \\'Gender\\' in this dataset is fine.\\n\")\n",
    "\n",
    "    return trait_biased, df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the JSON file into a DataFrame\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df\n",
    "\n",
    "\n",
    "def read_json_to_dataframe(json_file: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads a JSON file and converts it into a pandas DataFrame.\n",
    "    \"\"\"\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    return pd.DataFrame.from_dict(data, orient='index').reset_index().rename(columns={'index': 'cohort_id'})\n",
    "\n",
    "\n",
    "def filter_and_rank_cohorts(json_file: str, condition: Union[str, None] = None) -> Tuple[\n",
    "    Union[str, None], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Reads a JSON file, filters cohorts based on usability and an optional condition, then ranks them by sample size.\n",
    "\n",
    "    Args:\n",
    "    json_file (str): The path to the JSON file containing the data.\n",
    "    condition (str, optional): An additional condition for filtering. If None, only 'is_usable' is considered.\n",
    "\n",
    "    Returns:\n",
    "    Tuple: A tuple containing the best cohort ID (str or None if no suitable cohort is found) and\n",
    "           the filtered and ranked DataFrame.\n",
    "    \"\"\"\n",
    "    df = read_json_to_dataframe(json_file)\n",
    "    if condition:\n",
    "        filtered_df = df[(df['is_usable'] == True) & (df[condition] == True)]\n",
    "    else:\n",
    "        filtered_df = df[df['is_usable'] == True]\n",
    "    ranked_df = filtered_df.sort_values(by='sample_size', ascending=False)\n",
    "    best_cohort_id = ranked_df.iloc[0]['cohort_id'] if not ranked_df.empty else None\n",
    "\n",
    "    return best_cohort_id, ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_batch_effect(X):\n",
    "    \"\"\"\n",
    "    Detect potential batch effects in a dataset using eigenvalues of XX^T.\n",
    "\n",
    "    Args:\n",
    "    X (numpy.ndarray): A feature matrix with shape (n_samples, n_features).\n",
    "\n",
    "    Returns:\n",
    "    bool: True if a potential batch effect is detected, False otherwise.\n",
    "    \"\"\"\n",
    "    n_samples = X.shape[0]\n",
    "\n",
    "    # Computing XX^T\n",
    "    XXt = np.dot(X, X.T)\n",
    "\n",
    "    # Compute the eigenvalues of XX^T\n",
    "    eigen_values = np.linalg.eigvalsh(XXt)  # Using eigvalsh since XX^T is symmetric\n",
    "    eigen_values = sorted(eigen_values, reverse=True)\n",
    "\n",
    "    # Check for large gaps in the eigenvalues\n",
    "    for i in range(len(eigen_values) - 1):\n",
    "        gap = eigen_values[i] - eigen_values[i + 1]\n",
    "        if gap > 1 / n_samples:  # You may need to adjust this threshold\n",
    "            return True\n",
    "\n",
    "    return False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_data(X_train, X_test=None):\n",
    "    \"\"\"Compute the mean and standard deviation statistics of the training data, use them to normalize the training data,\n",
    "    and optionally the test data\"\"\"\n",
    "    mean = np.mean(X_train, axis=0)\n",
    "    std = np.std(X_train, axis=0)\n",
    "\n",
    "    # Handling columns with std = 0\n",
    "    std_no_zero = np.where(std == 0, 1, std)\n",
    "\n",
    "    # Normalize X_train\n",
    "    X_train_normalized = (X_train - mean) / std_no_zero\n",
    "    # Set normalized values to 0 where std was 0\n",
    "    X_train_normalized[:, std == 0] = 0\n",
    "\n",
    "    if X_test is not None:\n",
    "        X_test_normalized = (X_test - mean) / std_no_zero\n",
    "        X_test_normalized[:, std == 0] = 0\n",
    "    else:\n",
    "        X_test_normalized = None\n",
    "\n",
    "    return X_train_normalized, X_test_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ResidualizationRegressor:\n",
    "    def __init__(self, regression_model_constructor, params=None):\n",
    "        if params is None:\n",
    "            params = {}\n",
    "        self.regression_model = regression_model_constructor(**params)\n",
    "        self.beta_Z = None  # Coefficients for regression of Y on Z\n",
    "        self.beta_X = None  # Coefficients for regression of residual on X\n",
    "        self.neg_log_p_values = None  # Negative logarithm of p-values\n",
    "        self.p_values = None  # Actual p-values\n",
    "\n",
    "    def _reshape_data(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the data to ensure it's in the correct format (2D array).\n",
    "\n",
    "        :param data: The input data (can be 1D or 2D array).\n",
    "        :return: Reshaped 2D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 1:\n",
    "            return data.reshape(-1, 1)\n",
    "        return data\n",
    "\n",
    "    def _reshape_output(self, data):\n",
    "        \"\"\"\n",
    "        Reshape the output data to ensure it's in the correct format (1D array).\n",
    "\n",
    "        :param data: The output data (can be 1D or 2D array).\n",
    "        :return: Reshaped 1D array.\n",
    "        \"\"\"\n",
    "        if data.ndim == 2 and data.shape[1] == 1:\n",
    "            return data.ravel()\n",
    "        return data\n",
    "\n",
    "    def fit(self, X, Y, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Y = self._reshape_data(Y)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        # Step 1: Linear regression of Y on Z\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        self.beta_Z = np.linalg.pinv(Z_ones.T @ Z_ones) @ Z_ones.T @ Y\n",
    "        Y_hat = Z_ones @ self.beta_Z\n",
    "        e_Y = Y - Y_hat  # Residual of Y\n",
    "\n",
    "        # Step 2: Regress the residual on X using the included regression model\n",
    "        self.regression_model.fit(X, e_Y)\n",
    "\n",
    "        # Obtain coefficients from the regression model\n",
    "        if hasattr(self.regression_model, 'coef_'):\n",
    "            self.beta_X = self.regression_model.coef_\n",
    "        elif hasattr(self.regression_model, 'getBeta'):\n",
    "            beta_output = self.regression_model.getBeta()\n",
    "            self.beta_X = self._reshape_output(beta_output)\n",
    "\n",
    "        # Obtain negative logarithm of p-values, if available\n",
    "        if hasattr(self.regression_model, 'getNegLogP'):\n",
    "            neg_log_p_output = self.regression_model.getNegLogP()\n",
    "            if neg_log_p_output is not None:\n",
    "                self.neg_log_p_values = self._reshape_output(neg_log_p_output)\n",
    "                self.p_values = np.exp(-self.neg_log_p_values)\n",
    "                # Concatenate the p-values of Z and X. The p-values of Z were not computed, mark with NaN.\n",
    "                p_values_Z = np.full(Z.shape[1], np.nan)\n",
    "                self.p_values = np.concatenate((p_values_Z, self.p_values))\n",
    "\n",
    "    def predict(self, X, Z):\n",
    "        X = self._reshape_data(X)\n",
    "        Z = self._reshape_data(Z)\n",
    "\n",
    "        Z_ones = np.column_stack((np.ones(Z.shape[0]), Z))\n",
    "        ZX = np.column_stack((Z, X))\n",
    "        combined_beta = np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "        return ZX @ combined_beta + self.beta_Z[0]\n",
    "\n",
    "    def get_coefficients(self):\n",
    "        return np.concatenate((self.beta_Z[1:].ravel(), self.beta_X.ravel()))\n",
    "\n",
    "    def get_p_values(self):\n",
    "        return self.p_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cross_validation(X, Y, Z, model_constructor, model_params, k=5, target_type='binary'):\n",
    "    assert target_type in ['binary', 'continuous'], \"The target type must be chosen from 'binary' or 'continuous'\"\n",
    "    indices = np.arange(X.shape[0])\n",
    "    np.random.shuffle(indices)\n",
    "\n",
    "    fold_size = len(X) // k\n",
    "    performances = []\n",
    "\n",
    "    for i in range(k):\n",
    "        # Split data into train and test based on the current fold\n",
    "        test_indices = indices[i * fold_size: (i + 1) * fold_size]\n",
    "        train_indices = np.setdiff1d(indices, test_indices)\n",
    "\n",
    "        X_train, X_test = X[train_indices], X[test_indices]\n",
    "        Y_train, Y_test = Y[train_indices], Y[test_indices]\n",
    "        Z_train, Z_test = Z[train_indices], Z[test_indices]\n",
    "\n",
    "        normalized_X_train, normalized_X_test = normalize_data(X_train, X_test)\n",
    "        normalized_Z_train, normalized_Z_test = normalize_data(Z_train, Z_test)\n",
    "\n",
    "        # model = model_constructor(**model_params)\n",
    "        model = ResidualizationRegressor(model_constructor, model_params)\n",
    "        model.fit(normalized_X_train, Y_train, normalized_Z_train)\n",
    "        predictions = model.predict(normalized_X_test, normalized_Z_test)\n",
    "\n",
    "        if target_type == 'binary':\n",
    "            predictions = (predictions > 0.5).astype(int)\n",
    "            Y_test = (Y_test > 0.5).astype(int)\n",
    "            performance = accuracy_score(Y_test, predictions)\n",
    "        elif target_type == 'continuous':\n",
    "            performance = mean_squared_error(Y_test, predictions)\n",
    "\n",
    "        performances.append(performance)\n",
    "\n",
    "    cv_mean = np.mean(performances)\n",
    "    cv_std = np.std(performances)\n",
    "\n",
    "    if target_type == 'binary':\n",
    "        print(f'The cross-validation accuracy is {(cv_mean * 100):.2f}% ± {(cv_std * 100):.2f}%')\n",
    "    else:\n",
    "        print(f'The cross-validation MSE is {(cv_mean * 100):.2f} ± {(cv_std * 100):.2f}')\n",
    "\n",
    "    return cv_mean, cv_std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def interpret_result(model: Any, feature_names: List[str], trait: str, condition: str,\n",
    "                     threshold: float = 0.05, save_output: bool = True,\n",
    "                     output_dir: str = './output', model_id: int = 1) -> None:\n",
    "    \"\"\"This function interprets and reports the result of a trained linear regression model, where the regressor\n",
    "    consists of one variable about condition and multiple variables about genetic factors.\n",
    "    The function extracts coefficients and p-values from the model, and identifies the significant genes based on\n",
    "    p-values or non-zero coefficients, depending on the availability of p-values.\n",
    "\n",
    "    Parameters:\n",
    "    model (Any): The trained regression Model.\n",
    "    feature_names (List[str]): A list of feature names corresponding to the model's coefficients.\n",
    "    trait (str): The target trait of interest.\n",
    "    condition (str): The specific condition to examine within the model.\n",
    "    threshold (float): Significance level for p-value correction. Defaults to 0.05.\n",
    "    save_output (bool): Flag to determine whether to save the output to a file. Defaults to True.\n",
    "    output_dir (str): Directory path where output files are saved. Defaults to './output'.\n",
    "    model_id (int): The index of the model, 1 or 2.\n",
    "\n",
    "    Returns:\n",
    "    None: This function does not return anything but prints and optionally saves the output.\n",
    "    \"\"\"\n",
    "    coefficients = model.get_coefficients().reshape(-1).tolist()\n",
    "    p_values = model.get_p_values()\n",
    "    if p_values is None:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients\n",
    "        })\n",
    "    else:\n",
    "        regression_df = pd.DataFrame({\n",
    "            'Variable': feature_names,\n",
    "            'Coefficient': coefficients,\n",
    "            'p_value': p_values.reshape(-1).tolist()\n",
    "        })\n",
    "\n",
    "    condition_effect = regression_df[regression_df['Variable'] == condition].iloc[0]\n",
    "\n",
    "    print(f\"Effect of the condition on the target variable:\")\n",
    "    print(f\"Variable: {condition}\")\n",
    "    print(f\"Coefficient: {condition_effect['Coefficient']:.4f}\")\n",
    "    gene_regression_df = regression_df[regression_df['Variable'] != condition]\n",
    "    if p_values is None:\n",
    "        gene_regression_df['Absolute Coefficient'] = gene_regression_df['Coefficient'].abs()\n",
    "        significant_genes = gene_regression_df[gene_regression_df['Coefficient'] != 0]\n",
    "        significant_genes_sorted = significant_genes.sort_values(by='Absolute Coefficient', ascending=False)\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} genes with non-zero coefficients associated with the trait '{trait}' \"\n",
    "            f\"conditional on the factor '{condition}'. These genes are identified as significant based on the regression model.\")\n",
    "    else:\n",
    "        # Apply the Benjamini-Hochberg correction, to get the corrected p-values\n",
    "        corrected_p_values = multipletests(gene_regression_df['p_value'], alpha=threshold, method='fdr_bh')[1]\n",
    "        gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n",
    "        significant_genes = gene_regression_df.loc[gene_regression_df['corrected_p_value'] < threshold]\n",
    "        significant_genes_sorted = significant_genes.sort_values('corrected_p_value')\n",
    "        print(\n",
    "            f\"Found {len(significant_genes_sorted)} significant genes associated with the trait '{trait}' conditional on \"\n",
    "            f\"the factor '{condition}', with corrected p-value < {threshold}:\")\n",
    "\n",
    "    print(significant_genes_sorted.to_string(index=False))\n",
    "\n",
    "    # Optionally, save this to a CSV file\n",
    "    if save_output:\n",
    "        significant_genes_sorted.to_csv(\n",
    "            os.path.join(output_dir, f'significant_genes_condition_{condition}_{model_id}.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:37:25.927828954Z",
     "start_time": "2024-01-10T21:37:25.311064788Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "\n",
    "from utils import *\n",
    "# Set your preferred name\n",
    "USER = \"Miantong\"\n",
    "# Set the data and output directories\n",
    "DATA_ROOT = '/Users/mz/Desktop/AI4Science/DATA/'\n",
    "OUTPUT_ROOT = '/Users/mz/Desktop/AI4Science/DATA/output1'\n",
    "TRAIT = 'Arrhythmia'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n",
    "\n",
    "# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n",
    "NORMALIZE_GENE = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:25:23.253882615Z",
     "start_time": "2023-12-31T03:25:23.244062710Z"
    },
    "collapsed": false
   },
   "source": [
    "In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:38:31.425592571Z",
     "start_time": "2024-01-10T21:38:31.419137625Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.DS_Store',\n",
       " 'GSE143924',\n",
       " 'GSE93101',\n",
       " 'GSE213029',\n",
       " 'GSE55231',\n",
       " 'GSE41177',\n",
       " 'GSE34788',\n",
       " 'GSE47727',\n",
       " 'GSE66032',\n",
       " 'GSE47728',\n",
       " 'GSE136992',\n",
       " 'GSE115574',\n",
       " 'GSE53622',\n",
       " 'GSE97455',\n",
       " 'GSE53624']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Arrhythmia\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/mz/Desktop/AI4Science/DATA/GEO/Arrhythmia/GSE213029/GSE213029_family.soft.gz',\n",
       " '/Users/mz/Desktop/AI4Science/DATA/GEO/Arrhythmia/GSE213029/GSE213029_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = accession_num = \"GSE34788\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE41177\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE47727\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE47728\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE53622\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE53624\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE55231\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE66032\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE93101\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE97455\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE115574\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE136992\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE143924\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file\n",
    "cohort = accession_num = \"GSE213029\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:03.718770427Z",
     "start_time": "2024-01-10T21:39:03.677006442Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Aerobic Exercise Training Rejuvenates the Human Skeletal Muscle Methylome Ten years after Breast Cancer Treatment and Survival\"\n",
      "!Series_summary\t\"Many cancer survivors suffer from impairments in skeletal muscle (SkM), both in terms of reduced mass and function. Interestingly, human SkM possesses an epigenetic memory of earlier stimuli, such as exercise. Long-term retention of epigenetic changes in SkM following cancer survival and/or exercise training have not yet been studied. We therefore investigated genome-wide DNA methylation (the methylome) in SkM following a 5-month, 3/week treadmill-based aerobic training intervention in breast cancer survivors 10-14 years after diagnosis and treatment. These results were compared to breast cancer survivors who remained untrained and to age-matched controls with no history of cancer, who undertook the same training intervention. SkM biopsies were obtained before (pre) and after (post) the 5-month training period and DNA methylation was analysed using InfiniumEPIC 850K Arrays. The breast cancer survivors displayed a significant retention of increased DNA methylation (i.e., hypermethylation) at a larger number of differentially methylated positions (DMPs) compared with healthy age-matched controls pre-training. Training in cancer survivors led to an exaggerated number of DMPs with a hypermethylated signature occurring at more random non-regulatory regions across the DNA compared with training in healthy age-matched controls. However, the opposite occurred in important gene regulatory regions, where training in cancer survivors elicited a considerable reduction in methylation (i.e., hypomethylation) in 99% of the DMPs located specifically in CpG islands within promoter regions. Importantly, training was able to reverse the hypermethylation identified in cancer survivors back towards a hypomethylated signature that was observed pre-training in healthy age-matched controls at 300 (out of 881) of these island and promoter associated CpG sites. Pathway enrichment analysis identified that the training in cancer survivors evoked this predominantly hypomethylated signature in pathways associated with: Cell cycle, DNA replication and repair, transcription, translation, mTOR signalling and proteosome. Differentially methylated region (DMR) analysis also identified genes: BAG1, BTG2, CHP1, KIFC1, MKL2, MTR, PEX11B, POLD2, S100A6, SNORD104 and SPG7 as hypermethylated in breast cancer survivors with training reversing these CpG island / promoter associated DMRs towards a hypomethylated signature. Training also elicited a largely different epigenetic response in healthy individuals than that observed in cancer survivors, with very few overlapping changes. Only one gene, SIRT2, was identified as having altered methylation in cancer survivors at baseline as well as after training in both the cancer survivors and healthy controls. In conclusion, human SkM muscle retains a hypermethylated signature for as long as 10-14 years after breast cancer treatment and survival. Importantly, 5 months of aerobic training rejuvenated the SkM methylome towards signatures identified in healthy age-matched individuals in gene regulatory regions.\"\n",
      "!Series_overall_design\t\"Ethical approval: \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"This study is part of the ongoing CAUSE-trial (Cardiovascular Survivor Exercise) initiated by the Norwegian School of Sport Sciences (NSSS) and Oslo University Hospital. The Regional Committees for Medical and Health Research Ethics (REK) approved the study (reference number 28930; 2019/1318), and the study was pre-registered in clinicaltrials.gov (NCT04307407). \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Participants and Experimental Design: \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"The CAUSE-trial is a two-armed randomised controlled exercise training trial, where breast cancer survivors are randomised to either an exercise group or a control group (usual care), that also includes third arm of age-matched women with no prior history of cancer undertaking the same exercise training program. Breast cancer survivors diagnosed, with stage II-III HER2 negative breast cancer at the age of 60 years or less, between 2008 and 2012 were identified by the Cancer Registry of Norway. The study included survivors that had received anthracycline-based chemotherapy (Epirubicin), were currently exercising less than 90 minutes at moderate to high intensity per week, and that were living within driving distance from NSSS. Major exclusion criteria for the study were stage IV breast cancer diagnosis, adjuvant Trastuzumab-treatment, recurrent breast cancer or presence of secondary cancers, previous major cardiac surgery, pacemaker, chronic atrial fibrillation or any recent or uncontrolled cardiovascular disease. Participants for the healthy age-matched control group were recruited via advertisements. The participants were matched for age with the breast cancer survivors and could not have a history of any cancer and otherwise had to adhere to the same eligibility criteria as the breast cancer survivors (except for the breast cancer diagnosis). \"\n",
      "!Series_overall_design\t\"   \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"The breast cancer survivors were block randomised with a 1:1 allocation ratio to either an aerobic exercise training group (Cancer Trained) or to a control group (Cancer Untrained) for 5 months (20 weeks). The training intervention is described below. Participants in the healthy age-matched group did not undergo randomization, but all underwent the same exercise protocol as the cancer trained group (Healthy Age-Matched Trained). All participant groups had SkM biopsies taken and underwent other study-related assessments pre and post training. SkM biopsy methods are detailed below.     Aerobic Exercise Training: Both the cancer survivors training group and the healthy age-matched group underwent five months of 3 times per week treadmill-based endurance training, aimed at increasing V̇O2peak. The program followed a non-linear periodization model, typically with one hard (i.e., 80-87% of HRpeak), one moderate (i.e., 70-80% of HRpeak) and one easy (i.e., 60-70% of HRpeak) session per week. Initially, all sessions were continuous in duration, but from week 5, the hard session was an interval session, with 3-5 intervals each lasting 4 minutes (i.e., 90-97% of HRpeak). After week 10, we introduced 8-minute intervals for the moderate session (i.e., 85-91% of HRpeak), progressing from 2-4 intervals throughout the intervention.     \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Skeletal Muscle Biopsies:   \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"The biopsy procedure was conducted under local anaesthesia (Xylocaine with adrenaline, 10 mg/ml lidocaine + 5 µg/ml adrenaline, AstraZeneca, London, UK) and muscle tissue was obtained using a modified Bergström technique with suction. The tissue was quickly rinsed in physiological saline, before fat, connective tissue and blood were removed and discarded. Subsequently, the samples were quickly frozen in isopentane and cooled on dry ice or liquid nitrogen, before being transferred and stored at −80°C for later isolation of DNA. All biopsies from both pre and post training time points were taken at rest, in the morning after an overnight fast from the lateral portion of vastus lateralis muscle. Post training samples were obtained at least 72 hrs after the end of the training intervention. Pre and post biopsies were taken from the left leg in all participants.   \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Cardiopulmonary Exercise Test and Physical Activity levels:   \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Participants body height and mass were recorded using a stadiometer and a scale (SECA 213, Hamburg, Germany). Cardiorespiratory fitness (V̇O2peak) was assessed by a treadmill-based, symptom-limited cardiopulmonary exercise test, using a modified Balke protocol. After familiarisation with the treadmill, participants were encouraged to walk at gradually increasing exercise loads (increased inclination or speed) until voluntary exhaustion. Continuous expired air was measured breath-by-breath, using a gas and volume calibrated metabolic chart (Oxycon Pro, Jarger GmbH, Hoechberg, Germany). Objectively measured physical activity levels were recorded for seven consecutive days, using the ActiGraph™ model GT3X+ (ActiGraph LLC, Pensacola FL, USA). The average minutes spent in light and moderate-to-vigorous physical activity, using cut-off values used previously [1], with light physical activity defined as 100-1951 counts per minute and moderate-to-vigorous physical activity >1952 counts per minute. Baseline characteristics of V̇O2peak and physical activity levels for all participants can be found below. There was no significant difference in V̇O2peak or physical activity levels between the cancer groups and the healthy age-matched group prior to the start of the exercise program.     \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Tissue Homogenization, DNA isolation, Bisulfite Conversion and Genome-Wide DNA Methylation:\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"DNA was isolated from SkM tissue derived from a randomly selected subpopulation within each group of the larger trial (n=5 cancer trained, n=5 cancer untrained and n=6 healthy age-matched trained) at both pre and post training. These subpopulations baseline characteristics can be found here in the order of: Cancer Trained (n=5), Cancer Untrained (n=5), Healthy Age-Matched Trained (n=6), Cancer Survivors vs. Healthy Age-matched (p-value) respectively for: Age (years; SD) 60.4 (5.9), 62.0 (5.5), 58.3 (4.8), p = 0.307; Body Height (cm; SD) 167.5 (5,9), 168.1 (5.0), 169.0 (4.8), p = 0.660; Body Mass (Kg; SD) 80.5 (7.9), 74.1 (8.6), 80.0 (16.5), p = 0.670; Body Mass Index (SD) 28.8 (3.8), 26.4 (4.3), 28.0 (5.8), p = 0.861; V̇O2peak (ml/kg/min; SD) 27.2 (3.9),30.9 (6.7), 30.7 (7.7), p = 0.630; Physical activity (PA) levels (minutes; SD) for Light PA 154 (44.9),154 (38.4), 162.9 (28.3), p = 0.652; Moderate to Vigorous PA 43.3 (34.6), 41.8 (16.4), 56.2 (31.6), p = 0.362. Baseline characteristics were well balanced between the groups. There was no difference in age, height, BMI, V̇O2peak or physical activity levels between the cancer groups and the healthy age-matched group. Muscle samples were homogenized for 45 seconds at 6,000 rpm × 3 (5 min on ice in-between intervals) in lysis buffer (180 µl buffer ATL with 20 µl proteinase K) provided in the DNeasy spin column kit (Qiagen, UK) using a Roche Magnalyser instrument and homogenization tubes containing ceramic beads (Roche, UK). The DNA was then bisulfite converted using the EZ DNA Methylation Kit (Zymo Research, CA, United States) as per the manufacturer’s instructions.    Infinium MethylationEPIC BeadChip Array    All DNA methylation experiments were performed in accordance with Illumina manufacturer instructions for the Infinium Methylation EPIC BeadChip Array. Methods for the amplification, fragmentation, precipitation and resuspension of amplified DNA, hybridisation to EPIC beadchip, extension and staining of the bisulfite converted DNA (BCD) can be found in detail in our open access methods paper [2, 3]. EPIC BeadChips were imaged using the Illumina iScan System (Illumina, United States).     \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"DNA methylome analysis, differentially methylated Positions (DMPs), pathway enrichment analysis (KEGG pathways) and differentially methylated region (DMR) analysis:    \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"Following MethylationEPIC BeadChip arrays, raw .IDAT files were processed using Partek Genomics Suite V.7 (Partek Inc. Missouri, USA) and annotated using the MethylationEPIC_v-1-0_B4 manifest file. The mean detection p-value for all samples was 0.0004, which was well below the recommended 0.01 in the Oshlack workflow [4]. We also examined the raw intensities/signals of the probes, that demonstrated an average median methylated and unmethylated signal of 11.77 and 11.69, respectively, with an average of 11.73, which is recommended to be above 11.5 [4]. The difference between the average median methylated and average median unmethylated signal was 0.07, well below the recommended difference of less than 0.5 [4]. Note: Upon import of the data into Partek Genomics Suite we filtered out probes located in known single nucleotide polymorphisms (SNPs) and any known cross-reactive probes using previously defined SNP and cross-reactive probe lists from EPIC BeadChip 850K validation studies [5]. Although the average detection p-value for each sample across all probes was very low (on average 0.0004) we also excluded any individual probes with a detection p-value that was above 0.01 as recommended previously [4]. Furthermore, due to an all-female cohort of breast cancer survivors and age-matched healthy controls, we filtered out any probes on the male only Y chromosome (therefore the analysis included all probes on the X chromosome only). Out of a total of 865,860 probes in the EPIC array, removal of known SNPs, cross-reactive probes, those with a detection p-value above 0.01 and those on the Y chromosome resulted in 808,804 probes being taken forward for downstream analysis. Following this, background normalisation was performed via functional normalisation (with noob background correction) as previously described [6]. After functional normalisation, we also undertook quality control procedures via principal component analysis (PCA). One sample (Sample No.13_205624860016_R05C01_Cancer Untrained_Pre_Participant 10)  was removed due to a larger variation than that expected within that condition (variation defined as values above 2.2 standard deviations (SDs) for that condition – depicted by ellipsoids in the PCA plots. Following normalisation and quality control procedures, we undertook differentially methylated position (DMP) analysis by converting β-values to M-values (M-value = log2(β / (1 - β)), as M-values show distributions that are more statistically valid for the differential analysis of methylation levels [7]. We then performed a two-way ANOVA for group/condition (cancer trained, cancer untrained, healthy age-matched trained) and time (pre and post), with planned contrast/pairwise comparisons of: Cancer Trained-Pre vs. Healthy Trained-Pre (to investigate the effect of Cancer on the methylome named - Cancer Alone), Cancer Trained-Post vs. Cancer Trained-Pre (to investigate the impact of training on the methylome of breast cancer survivors named - Training in Cancer), Healthy Age-Matched Trained-Post vs. Heathy Age-Matched-Pre (to investigate the impact of training on the methylome in age-matched healthy controls named- Training in Healthy), Cancer Untrained-Post vs. Cancer Untrained-Pre (to help assess if any methylation changes occur over the time of the intervention period in cancer survivors, named - Experimental Time), and finally Cancer Trained Pre vs. Cancer Untrained Pre (to assess if there were any differences in methylation between the cancer groups at baseline, named - Between Cancer Groups at Baseline). For initial discovery of CpG sites that were deemed statistically significant, DMPs with an unadjusted P value of ≤ 0.01 were accepted for downstream analysis (Kyoto Encyclopedia of Genes and Genomes/KEGG pathway, differentially methylated region/DMR analysis - see below). We then undertook CpG enrichment analysis on these DMPs within KEGG pathways [8-10] using Partek Genomics Suite and Partek Pathway software. DMR analysis was performed to identify where several CpGs were differentially methylated within a short chromosomal locations/regions, undertaken using the Bioconductor package DMRcate (DOI: 10.18129/B9.bioc.DMRcate).     \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"IMPORTANT NOTE for Methylation Array files:  \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"For submission of data to GEO it is recommended to upload all normalized data for ALL samples and ALL probes. Therefore, for the matrix signal intensities we include normalized data for all probes and all samples as the GEO submission. Therefore, it’s important to note, (as described in detail in the methods above), sample no. 13 (Sample No.13_205624860016_R05C01_Cancer Untrained_Pre_Participant 10) was removed from downstream analysis. Also, for any downstream analysis (DMP, DMR, GO / KEGG analysis etc.) we removed cross-reactive probes, probes on SNPs, probes with a detection p value of above 0.01 and due to an all-female cohort, we also removed any probes on the male only Y chromosome (therefore including all probes on the X chromosome only). We also removed 378 DMPs from downstream analysis that included those that were altered over the experimental time period (5 months) in the cancer untrained group as well as those that were significantly different between cancer groups at baseline (specific details can be found in the final publication).\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"REFERENCES:   \"\n",
      "!Series_overall_design\t\"\"\n",
      "!Series_overall_design\t\"1. Ekelund U, Tarp J, Fagerland MW, Johannessen JS, Hansen BH, Jefferis BJ, et al. Joint associations of accelero-meter measured physical activity and sedentary time with all-cause mortality: a harmonised meta-analysis in more than 44 000 middle-aged and older individuals. Br J Sports Med. 2020; 54(24):1499-1506.  \"\n",
      "!Series_overall_design\t\"2. Seaborne RA, Strauss J, Cocks M, Shepherd S, O'Brien TD, Someren KAV, et al. Methylome of human skeletal muscle after acute & chronic resistance exercise training, detraining & retraining. Scientific Data (Nature). 2018; 5:180213.  \"\n",
      "!Series_overall_design\t\"3. Seaborne RA, Strauss J, Cocks M, Shepherd S, O'Brien TD, van Someren KA, et al. Human Skeletal Muscle Possesses an Epigenetic Memory of Hypertrophy. Scientific Reports (Nature) 2018; 8(1):1898.  \"\n",
      "!Series_overall_design\t\"4. Maksimovic J, Phipson B, and Oshlack A. A cross-package Bioconductor workflow for analysing methylation array data [version 1; referees: 3 approved, 1 approved with reservations]. F1000Research. 2016; 5(1281).  \"\n",
      "!Series_overall_design\t\"5. Pidsley R, Zotenko E, Peters TJ, Lawrence MG, Risbridger GP, Molloy P, et al. Critical evaluation of the Illumina MethylationEPIC BeadChip microarray for whole-genome DNA methylation profiling. Genome Biology. 2016; 17(1):208.  \"\n",
      "!Series_overall_design\t\"6. Maksimovic J, Gordon L, and Oshlack A. SWAN: Subset-quantile within array normalization for illumina infinium HumanMethylation450 BeadChips. Genome Biol. 2012; 13(6):R44.  \"\n",
      "!Series_overall_design\t\"7. Du P, Zhang X, Huang C-C, Jafari N, Kibbe WA, Hou L, et al. Comparison of Beta-value and M-value methods for quantifying methylation levels by microarray analysis. BMC Bioinformatics. 2010; 11(1):587.  \"\n",
      "!Series_overall_design\t\"8. Kanehisa M and Goto S. KEGG: kyoto encyclopedia of genes and genomes. Nucleic Acids Res. 2000; 28(1):27-30.  \"\n",
      "!Series_overall_design\t\"9. Kanehisa M, Furumichi M, Tanabe M, Sato Y, and Morishima K. KEGG: new perspectives on genomes, pathways, diseases and drugs. Nucleic Acids Res. 2017; 45(D1):D353-d361.  \"\n",
      "!Series_overall_design\t\"10. Kanehisa M, Sato Y, Kawashima M, Furumichi M, and Tanabe M. KEGG as a reference resource for gene and protein annotation. Nucleic Acids Res. 2016; 44(D1):D457-62.\"\n"
     ]
    }
   ],
   "source": [
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.022548957Z",
     "start_time": "2024-01-10T21:39:04.008519515Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM6569559</th>\n",
       "      <th>GSM6569560</th>\n",
       "      <th>GSM6569561</th>\n",
       "      <th>GSM6569562</th>\n",
       "      <th>GSM6569563</th>\n",
       "      <th>GSM6569564</th>\n",
       "      <th>GSM6569565</th>\n",
       "      <th>GSM6569566</th>\n",
       "      <th>GSM6569567</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM6569581</th>\n",
       "      <th>GSM6569582</th>\n",
       "      <th>GSM6569583</th>\n",
       "      <th>GSM6569584</th>\n",
       "      <th>GSM6569585</th>\n",
       "      <th>GSM6569586</th>\n",
       "      <th>GSM6569587</th>\n",
       "      <th>GSM6569588</th>\n",
       "      <th>GSM6569589</th>\n",
       "      <th>GSM6569590</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>gender: Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "      <td>tissue: Skeletal Muscle_Vastus Lateralis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                GSM6569559  \\\n",
       "0  !Sample_characteristics_ch1                            gender: Female   \n",
       "1  !Sample_characteristics_ch1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569560  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569561  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569562  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569563  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569564  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569565  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569566  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569567  ...  \\\n",
       "0                            gender: Female  ...   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis  ...   \n",
       "\n",
       "                                 GSM6569581  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569582  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569583  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569584  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569585  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569586  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569587  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569588  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569589  \\\n",
       "0                            gender: Female   \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis   \n",
       "\n",
       "                                 GSM6569590  \n",
       "0                            gender: Female  \n",
       "1  tissue: Skeletal Muscle_Vastus Lateralis  \n",
       "\n",
       "[2 rows x 33 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.307639583Z",
     "start_time": "2024-01-10T21:39:04.289352075Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: ['gender: Female'], 1: ['tissue: Skeletal Muscle_Vastus Lateralis']}"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data_unique = get_unique_values_by_row(clinical_data)\n",
    "clinical_data_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978204446Z",
     "start_time": "2023-12-31T03:58:04.922270095Z"
    },
    "collapsed": false
   },
   "source": [
    "Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:04.854586192Z",
     "start_time": "2024-01-10T21:39:04.836283516Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Arrhythmia\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Arrhythmia\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Aerobic Exercise Training Rejuvenates the Human Skeletal Muscle Methylome Ten years after Breast Cancer Treatment and Survival\"\\n!Series_summary\\t\"Many cancer survivors suffer from impairments in skeletal muscle (SkM), both in terms of reduced mass and function. Interestingly, human SkM possesses an epigenetic memory of earlier stimuli, such as exercise. Long-term retention of epigenetic changes in SkM following cancer survival and/or exercise training have not yet been studied. We therefore investigated genome-wide DNA methylation (the methylome) in SkM following a 5-month, 3/week treadmill-based aerobic training intervention in breast cancer survivors 10-14 years after diagnosis and treatment. These results were compared to breast cancer survivors who remained untrained and to age-matched controls with no history of cancer, who undertook the same training intervention. SkM biopsies were obtained before (pre) and after (post) the 5-month training period and DNA methylation was analysed using InfiniumEPIC 850K Arrays. The breast cancer survivors displayed a significant retention of increased DNA methylation (i.e., hypermethylation) at a larger number of differentially methylated positions (DMPs) compared with healthy age-matched controls pre-training. Training in cancer survivors led to an exaggerated number of DMPs with a hypermethylated signature occurring at more random non-regulatory regions across the DNA compared with training in healthy age-matched controls. However, the opposite occurred in important gene regulatory regions, where training in cancer survivors elicited a considerable reduction in methylation (i.e., hypomethylation) in 99% of the DMPs located specifically in CpG islands within promoter regions. Importantly, training was able to reverse the hypermethylation identified in cancer survivors back towards a hypomethylated signature that was observed pre-training in healthy age-matched controls at 300 (out of 881) of these island and promoter associated CpG sites. Pathway enrichment analysis identified that the training in cancer survivors evoked this predominantly hypomethylated signature in pathways associated with: Cell cycle, DNA replication and repair, transcription, translation, mTOR signalling and proteosome. Differentially methylated region (DMR) analysis also identified genes: BAG1, BTG2, CHP1, KIFC1, MKL2, MTR, PEX11B, POLD2, S100A6, SNORD104 and SPG7 as hypermethylated in breast cancer survivors with training reversing these CpG island / promoter associated DMRs towards a hypomethylated signature. Training also elicited a largely different epigenetic response in healthy individuals than that observed in cancer survivors, with very few overlapping changes. Only one gene, SIRT2, was identified as having altered methylation in cancer survivors at baseline as well as after training in both the cancer survivors and healthy controls. In conclusion, human SkM muscle retains a hypermethylated signature for as long as 10-14 years after breast cancer treatment and survival. Importantly, 5 months of aerobic training rejuvenated the SkM methylome towards signatures identified in healthy age-matched individuals in gene regulatory regions.\"\\n!Series_overall_design\\t\"Ethical approval: \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"This study is part of the ongoing CAUSE-trial (Cardiovascular Survivor Exercise) initiated by the Norwegian School of Sport Sciences (NSSS) and Oslo University Hospital. The Regional Committees for Medical and Health Research Ethics (REK) approved the study (reference number 28930; 2019/1318), and the study was pre-registered in clinicaltrials.gov (NCT04307407). \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Participants and Experimental Design: \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"The CAUSE-trial is a two-armed randomised controlled exercise training trial, where breast cancer survivors are randomised to either an exercise group or a control group (usual care), that also includes third arm of age-matched women with no prior history of cancer undertaking the same exercise training program. Breast cancer survivors diagnosed, with stage II-III HER2 negative breast cancer at the age of 60 years or less, between 2008 and 2012 were identified by the Cancer Registry of Norway. The study included survivors that had received anthracycline-based chemotherapy (Epirubicin), were currently exercising less than 90 minutes at moderate to high intensity per week, and that were living within driving distance from NSSS. Major exclusion criteria for the study were stage IV breast cancer diagnosis, adjuvant Trastuzumab-treatment, recurrent breast cancer or presence of secondary cancers, previous major cardiac surgery, pacemaker, chronic atrial fibrillation or any recent or uncontrolled cardiovascular disease. Participants for the healthy age-matched control group were recruited via advertisements. The participants were matched for age with the breast cancer survivors and could not have a history of any cancer and otherwise had to adhere to the same eligibility criteria as the breast cancer survivors (except for the breast cancer diagnosis). \"\\n!Series_overall_design\\t\"   \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"The breast cancer survivors were block randomised with a 1:1 allocation ratio to either an aerobic exercise training group (Cancer Trained) or to a control group (Cancer Untrained) for 5 months (20 weeks). The training intervention is described below. Participants in the healthy age-matched group did not undergo randomization, but all underwent the same exercise protocol as the cancer trained group (Healthy Age-Matched Trained). All participant groups had SkM biopsies taken and underwent other study-related assessments pre and post training. SkM biopsy methods are detailed below.     Aerobic Exercise Training: Both the cancer survivors training group and the healthy age-matched group underwent five months of 3 times per week treadmill-based endurance training, aimed at increasing V̇O2peak. The program followed a non-linear periodization model, typically with one hard (i.e., 80-87% of HRpeak), one moderate (i.e., 70-80% of HRpeak) and one easy (i.e., 60-70% of HRpeak) session per week. Initially, all sessions were continuous in duration, but from week 5, the hard session was an interval session, with 3-5 intervals each lasting 4 minutes (i.e., 90-97% of HRpeak). After week 10, we introduced 8-minute intervals for the moderate session (i.e., 85-91% of HRpeak), progressing from 2-4 intervals throughout the intervention.     \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Skeletal Muscle Biopsies:   \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"The biopsy procedure was conducted under local anaesthesia (Xylocaine with adrenaline, 10 mg/ml lidocaine + 5 µg/ml adrenaline, AstraZeneca, London, UK) and muscle tissue was obtained using a modified Bergström technique with suction. The tissue was quickly rinsed in physiological saline, before fat, connective tissue and blood were removed and discarded. Subsequently, the samples were quickly frozen in isopentane and cooled on dry ice or liquid nitrogen, before being transferred and stored at −80°C for later isolation of DNA. All biopsies from both pre and post training time points were taken at rest, in the morning after an overnight fast from the lateral portion of vastus lateralis muscle. Post training samples were obtained at least 72 hrs after the end of the training intervention. Pre and post biopsies were taken from the left leg in all participants.   \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Cardiopulmonary Exercise Test and Physical Activity levels:   \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Participants body height and mass were recorded using a stadiometer and a scale (SECA 213, Hamburg, Germany). Cardiorespiratory fitness (V̇O2peak) was assessed by a treadmill-based, symptom-limited cardiopulmonary exercise test, using a modified Balke protocol. After familiarisation with the treadmill, participants were encouraged to walk at gradually increasing exercise loads (increased inclination or speed) until voluntary exhaustion. Continuous expired air was measured breath-by-breath, using a gas and volume calibrated metabolic chart (Oxycon Pro, Jarger GmbH, Hoechberg, Germany). Objectively measured physical activity levels were recorded for seven consecutive days, using the ActiGraph™ model GT3X+ (ActiGraph LLC, Pensacola FL, USA). The average minutes spent in light and moderate-to-vigorous physical activity, using cut-off values used previously [1], with light physical activity defined as 100-1951 counts per minute and moderate-to-vigorous physical activity >1952 counts per minute. Baseline characteristics of V̇O2peak and physical activity levels for all participants can be found below. There was no significant difference in V̇O2peak or physical activity levels between the cancer groups and the healthy age-matched group prior to the start of the exercise program.     \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Tissue Homogenization, DNA isolation, Bisulfite Conversion and Genome-Wide DNA Methylation:\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"DNA was isolated from SkM tissue derived from a randomly selected subpopulation within each group of the larger trial (n=5 cancer trained, n=5 cancer untrained and n=6 healthy age-matched trained) at both pre and post training. These subpopulations baseline characteristics can be found here in the order of: Cancer Trained (n=5), Cancer Untrained (n=5), Healthy Age-Matched Trained (n=6), Cancer Survivors vs. Healthy Age-matched (p-value) respectively for: Age (years; SD) 60.4 (5.9), 62.0 (5.5), 58.3 (4.8), p = 0.307; Body Height (cm; SD) 167.5 (5,9), 168.1 (5.0), 169.0 (4.8), p = 0.660; Body Mass (Kg; SD) 80.5 (7.9), 74.1 (8.6), 80.0 (16.5), p = 0.670; Body Mass Index (SD) 28.8 (3.8), 26.4 (4.3), 28.0 (5.8), p = 0.861; V̇O2peak (ml/kg/min; SD) 27.2 (3.9),30.9 (6.7), 30.7 (7.7), p = 0.630; Physical activity (PA) levels (minutes; SD) for Light PA 154 (44.9),154 (38.4), 162.9 (28.3), p = 0.652; Moderate to Vigorous PA 43.3 (34.6), 41.8 (16.4), 56.2 (31.6), p = 0.362. Baseline characteristics were well balanced between the groups. There was no difference in age, height, BMI, V̇O2peak or physical activity levels between the cancer groups and the healthy age-matched group. Muscle samples were homogenized for 45 seconds at 6,000 rpm × 3 (5 min on ice in-between intervals) in lysis buffer (180 µl buffer ATL with 20 µl proteinase K) provided in the DNeasy spin column kit (Qiagen, UK) using a Roche Magnalyser instrument and homogenization tubes containing ceramic beads (Roche, UK). The DNA was then bisulfite converted using the EZ DNA Methylation Kit (Zymo Research, CA, United States) as per the manufacturer’s instructions.    Infinium MethylationEPIC BeadChip Array    All DNA methylation experiments were performed in accordance with Illumina manufacturer instructions for the Infinium Methylation EPIC BeadChip Array. Methods for the amplification, fragmentation, precipitation and resuspension of amplified DNA, hybridisation to EPIC beadchip, extension and staining of the bisulfite converted DNA (BCD) can be found in detail in our open access methods paper [2, 3]. EPIC BeadChips were imaged using the Illumina iScan System (Illumina, United States).     \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"DNA methylome analysis, differentially methylated Positions (DMPs), pathway enrichment analysis (KEGG pathways) and differentially methylated region (DMR) analysis:    \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Following MethylationEPIC BeadChip arrays, raw .IDAT files were processed using Partek Genomics Suite V.7 (Partek Inc. Missouri, USA) and annotated using the MethylationEPIC_v-1-0_B4 manifest file. The mean detection p-value for all samples was 0.0004, which was well below the recommended 0.01 in the Oshlack workflow [4]. We also examined the raw intensities/signals of the probes, that demonstrated an average median methylated and unmethylated signal of 11.77 and 11.69, respectively, with an average of 11.73, which is recommended to be above 11.5 [4]. The difference between the average median methylated and average median unmethylated signal was 0.07, well below the recommended difference of less than 0.5 [4]. Note: Upon import of the data into Partek Genomics Suite we filtered out probes located in known single nucleotide polymorphisms (SNPs) and any known cross-reactive probes using previously defined SNP and cross-reactive probe lists from EPIC BeadChip 850K validation studies [5]. Although the average detection p-value for each sample across all probes was very low (on average 0.0004) we also excluded any individual probes with a detection p-value that was above 0.01 as recommended previously [4]. Furthermore, due to an all-female cohort of breast cancer survivors and age-matched healthy controls, we filtered out any probes on the male only Y chromosome (therefore the analysis included all probes on the X chromosome only). Out of a total of 865,860 probes in the EPIC array, removal of known SNPs, cross-reactive probes, those with a detection p-value above 0.01 and those on the Y chromosome resulted in 808,804 probes being taken forward for downstream analysis. Following this, background normalisation was performed via functional normalisation (with noob background correction) as previously described [6]. After functional normalisation, we also undertook quality control procedures via principal component analysis (PCA). One sample (Sample No.13_205624860016_R05C01_Cancer Untrained_Pre_Participant 10)  was removed due to a larger variation than that expected within that condition (variation defined as values above 2.2 standard deviations (SDs) for that condition – depicted by ellipsoids in the PCA plots. Following normalisation and quality control procedures, we undertook differentially methylated position (DMP) analysis by converting β-values to M-values (M-value = log2(β / (1 - β)), as M-values show distributions that are more statistically valid for the differential analysis of methylation levels [7]. We then performed a two-way ANOVA for group/condition (cancer trained, cancer untrained, healthy age-matched trained) and time (pre and post), with planned contrast/pairwise comparisons of: Cancer Trained-Pre vs. Healthy Trained-Pre (to investigate the effect of Cancer on the methylome named - Cancer Alone), Cancer Trained-Post vs. Cancer Trained-Pre (to investigate the impact of training on the methylome of breast cancer survivors named - Training in Cancer), Healthy Age-Matched Trained-Post vs. Heathy Age-Matched-Pre (to investigate the impact of training on the methylome in age-matched healthy controls named- Training in Healthy), Cancer Untrained-Post vs. Cancer Untrained-Pre (to help assess if any methylation changes occur over the time of the intervention period in cancer survivors, named - Experimental Time), and finally Cancer Trained Pre vs. Cancer Untrained Pre (to assess if there were any differences in methylation between the cancer groups at baseline, named - Between Cancer Groups at Baseline). For initial discovery of CpG sites that were deemed statistically significant, DMPs with an unadjusted P value of ≤ 0.01 were accepted for downstream analysis (Kyoto Encyclopedia of Genes and Genomes/KEGG pathway, differentially methylated region/DMR analysis - see below). We then undertook CpG enrichment analysis on these DMPs within KEGG pathways [8-10] using Partek Genomics Suite and Partek Pathway software. DMR analysis was performed to identify where several CpGs were differentially methylated within a short chromosomal locations/regions, undertaken using the Bioconductor package DMRcate (DOI: 10.18129/B9.bioc.DMRcate).     \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"IMPORTANT NOTE for Methylation Array files:  \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"For submission of data to GEO it is recommended to upload all normalized data for ALL samples and ALL probes. Therefore, for the matrix signal intensities we include normalized data for all probes and all samples as the GEO submission. Therefore, it’s important to note, (as described in detail in the methods above), sample no. 13 (Sample No.13_205624860016_R05C01_Cancer Untrained_Pre_Participant 10) was removed from downstream analysis. Also, for any downstream analysis (DMP, DMR, GO / KEGG analysis etc.) we removed cross-reactive probes, probes on SNPs, probes with a detection p value of above 0.01 and due to an all-female cohort, we also removed any probes on the male only Y chromosome (therefore including all probes on the X chromosome only). We also removed 378 DMPs from downstream analysis that included those that were altered over the experimental time period (5 months) in the cancer untrained group as well as those that were significantly different between cancer groups at baseline (specific details can be found in the final publication).\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"REFERENCES:   \"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"1. Ekelund U, Tarp J, Fagerland MW, Johannessen JS, Hansen BH, Jefferis BJ, et al. Joint associations of accelero-meter measured physical activity and sedentary time with all-cause mortality: a harmonised meta-analysis in more than 44 000 middle-aged and older individuals. Br J Sports Med. 2020; 54(24):1499-1506.  \"\\n!Series_overall_design\\t\"2. Seaborne RA, Strauss J, Cocks M, Shepherd S, O\\'Brien TD, Someren KAV, et al. Methylome of human skeletal muscle after acute & chronic resistance exercise training, detraining & retraining. Scientific Data (Nature). 2018; 5:180213.  \"\\n!Series_overall_design\\t\"3. Seaborne RA, Strauss J, Cocks M, Shepherd S, O\\'Brien TD, van Someren KA, et al. Human Skeletal Muscle Possesses an Epigenetic Memory of Hypertrophy. Scientific Reports (Nature) 2018; 8(1):1898.  \"\\n!Series_overall_design\\t\"4. Maksimovic J, Phipson B, and Oshlack A. A cross-package Bioconductor workflow for analysing methylation array data [version 1; referees: 3 approved, 1 approved with reservations]. F1000Research. 2016; 5(1281).  \"\\n!Series_overall_design\\t\"5. Pidsley R, Zotenko E, Peters TJ, Lawrence MG, Risbridger GP, Molloy P, et al. Critical evaluation of the Illumina MethylationEPIC BeadChip microarray for whole-genome DNA methylation profiling. Genome Biology. 2016; 17(1):208.  \"\\n!Series_overall_design\\t\"6. Maksimovic J, Gordon L, and Oshlack A. SWAN: Subset-quantile within array normalization for illumina infinium HumanMethylation450 BeadChips. Genome Biol. 2012; 13(6):R44.  \"\\n!Series_overall_design\\t\"7. Du P, Zhang X, Huang C-C, Jafari N, Kibbe WA, Hou L, et al. Comparison of Beta-value and M-value methods for quantifying methylation levels by microarray analysis. BMC Bioinformatics. 2010; 11(1):587.  \"\\n!Series_overall_design\\t\"8. Kanehisa M and Goto S. KEGG: kyoto encyclopedia of genes and genomes. Nucleic Acids Res. 2000; 28(1):27-30.  \"\\n!Series_overall_design\\t\"9. Kanehisa M, Furumichi M, Tanabe M, Sato Y, and Morishima K. KEGG: new perspectives on genomes, pathways, diseases and drugs. Nucleic Acids Res. 2017; 45(D1):D353-d361.  \"\\n!Series_overall_design\\t\"10. Kanehisa M, Sato Y, Kawashima M, Furumichi M, and Tanabe M. KEGG as a reference resource for gene and protein annotation. Nucleic Acids Res. 2016; 44(D1):D457-62.\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'gender: Female\\'], 1: [\\'tissue: Skeletal Muscle_Vastus Lateralis\\']}\\n'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n",
    "1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n",
    "2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n",
    "   (1) Is there human data available for this trait?\n",
    "   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n",
    "   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n",
    "   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n",
    "\n",
    "Background information about the dataset:\n",
    "{background_info}\n",
    "\n",
    "Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n",
    "{clinical_data_unique}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978240504Z",
     "start_time": "2023-12-31T03:58:04.922365324Z"
    },
    "collapsed": false
   },
   "source": [
    "Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n",
    "Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:39:05.366304590Z",
     "start_time": "2024-01-10T21:39:05.356983082Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_row = gender_row = None\n",
    "convert_age = convert_gender = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.324053151Z",
     "start_time": "2024-01-10T21:40:09.251131090Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_gene_availabe = True\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = 0\n",
    "\n",
    "trait_type = 'binary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:09.912115984Z",
     "start_time": "2024-01-10T21:40:09.907913460Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "is_available = is_gene_availabe and (trait_row is not None)\n",
    "if not is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:10.982148317Z",
     "start_time": "2024-01-10T21:40:10.967368918Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 1)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def convert_trait(tissue_type):\n",
    "    \"\"\"\n",
    "    Convert tissue type to epilepsy presence (binary).\n",
    "    Assuming epilepsy presence for 'Hippocampus' tissue.\n",
    "    \"\"\"\n",
    "    if tissue_type == 'tissue: Skeletal Muscle_Vastus Lateralis':\n",
    "        return 1  \n",
    "    else:\n",
    "        return 0  \n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    \"\"\"\n",
    "    Convert gender string to a binary representation.\n",
    "    Female -> 0, Male -> 1, Unknown -> None\n",
    "    \"\"\"\n",
    "    if gender_string == 'gender: Female':\n",
    "        return 0\n",
    "    elif gender_string == 'gender: Male':\n",
    "        return 1\n",
    "    else:\n",
    "        return None  # In case of unknown gender\n",
    "\n",
    "# Example usage\n",
    "example_tissue = 'tissue: Skeletal Muscle_Vastus Lateralis'\n",
    "example_gender = 'gender: Male'\n",
    "\n",
    "converted_trait = convert_trait(example_tissue)\n",
    "converted_gender = convert_gender(example_gender)\n",
    "\n",
    "converted_trait, converted_gender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:11.684556802Z",
     "start_time": "2024-01-10T21:40:11.665079581Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM6569559</th>\n",
       "      <th>GSM6569560</th>\n",
       "      <th>GSM6569561</th>\n",
       "      <th>GSM6569562</th>\n",
       "      <th>GSM6569563</th>\n",
       "      <th>GSM6569564</th>\n",
       "      <th>GSM6569565</th>\n",
       "      <th>GSM6569566</th>\n",
       "      <th>GSM6569567</th>\n",
       "      <th>GSM6569568</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM6569581</th>\n",
       "      <th>GSM6569582</th>\n",
       "      <th>GSM6569583</th>\n",
       "      <th>GSM6569584</th>\n",
       "      <th>GSM6569585</th>\n",
       "      <th>GSM6569586</th>\n",
       "      <th>GSM6569587</th>\n",
       "      <th>GSM6569588</th>\n",
       "      <th>GSM6569589</th>\n",
       "      <th>GSM6569590</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Arrhythmia</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gender</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM6569559  GSM6569560  GSM6569561  GSM6569562  GSM6569563  \\\n",
       "Arrhythmia           1           1           1           1           1   \n",
       "Gender               0           0           0           0           0   \n",
       "\n",
       "            GSM6569564  GSM6569565  GSM6569566  GSM6569567  GSM6569568  ...  \\\n",
       "Arrhythmia           1           1           1           1           1  ...   \n",
       "Gender               0           0           0           0           0  ...   \n",
       "\n",
       "            GSM6569581  GSM6569582  GSM6569583  GSM6569584  GSM6569585  \\\n",
       "Arrhythmia           1           1           1           1           1   \n",
       "Gender               0           0           0           0           0   \n",
       "\n",
       "            GSM6569586  GSM6569587  GSM6569588  GSM6569589  GSM6569590  \n",
       "Arrhythmia           1           1           1           1           1  \n",
       "Gender               0           0           0           0           0  \n",
       "\n",
       "[2 rows x 32 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:13.509252107Z",
     "start_time": "2024-01-10T21:40:13.468235630Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM6569559</th>\n",
       "      <th>GSM6569560</th>\n",
       "      <th>GSM6569561</th>\n",
       "      <th>GSM6569562</th>\n",
       "      <th>GSM6569563</th>\n",
       "      <th>GSM6569564</th>\n",
       "      <th>GSM6569565</th>\n",
       "      <th>GSM6569566</th>\n",
       "      <th>GSM6569567</th>\n",
       "      <th>GSM6569568</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM6569581</th>\n",
       "      <th>GSM6569582</th>\n",
       "      <th>GSM6569583</th>\n",
       "      <th>GSM6569584</th>\n",
       "      <th>GSM6569585</th>\n",
       "      <th>GSM6569586</th>\n",
       "      <th>GSM6569587</th>\n",
       "      <th>GSM6569588</th>\n",
       "      <th>GSM6569589</th>\n",
       "      <th>GSM6569590</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000029</th>\n",
       "      <td>0.296613</td>\n",
       "      <td>0.262710</td>\n",
       "      <td>0.328604</td>\n",
       "      <td>0.244043</td>\n",
       "      <td>0.305002</td>\n",
       "      <td>0.351313</td>\n",
       "      <td>0.301256</td>\n",
       "      <td>0.223671</td>\n",
       "      <td>0.366115</td>\n",
       "      <td>0.352855</td>\n",
       "      <td>...</td>\n",
       "      <td>0.371031</td>\n",
       "      <td>0.308602</td>\n",
       "      <td>0.329137</td>\n",
       "      <td>0.309973</td>\n",
       "      <td>0.273167</td>\n",
       "      <td>0.377755</td>\n",
       "      <td>0.266383</td>\n",
       "      <td>0.322724</td>\n",
       "      <td>0.287259</td>\n",
       "      <td>0.395006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000103</th>\n",
       "      <td>0.494243</td>\n",
       "      <td>0.507445</td>\n",
       "      <td>0.559102</td>\n",
       "      <td>0.496870</td>\n",
       "      <td>0.617299</td>\n",
       "      <td>0.567107</td>\n",
       "      <td>0.678045</td>\n",
       "      <td>0.555238</td>\n",
       "      <td>0.560232</td>\n",
       "      <td>0.524053</td>\n",
       "      <td>...</td>\n",
       "      <td>0.621956</td>\n",
       "      <td>0.543419</td>\n",
       "      <td>0.558046</td>\n",
       "      <td>0.616681</td>\n",
       "      <td>0.618169</td>\n",
       "      <td>0.586974</td>\n",
       "      <td>0.568257</td>\n",
       "      <td>0.627627</td>\n",
       "      <td>0.562612</td>\n",
       "      <td>0.662220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000109</th>\n",
       "      <td>0.864064</td>\n",
       "      <td>0.816083</td>\n",
       "      <td>0.846933</td>\n",
       "      <td>0.821823</td>\n",
       "      <td>0.838450</td>\n",
       "      <td>0.825361</td>\n",
       "      <td>0.885593</td>\n",
       "      <td>0.862300</td>\n",
       "      <td>0.820759</td>\n",
       "      <td>0.851260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.832698</td>\n",
       "      <td>0.831781</td>\n",
       "      <td>0.789656</td>\n",
       "      <td>0.817778</td>\n",
       "      <td>0.805174</td>\n",
       "      <td>0.841091</td>\n",
       "      <td>0.806922</td>\n",
       "      <td>0.854048</td>\n",
       "      <td>0.799516</td>\n",
       "      <td>0.810282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000155</th>\n",
       "      <td>0.944174</td>\n",
       "      <td>0.944075</td>\n",
       "      <td>0.930635</td>\n",
       "      <td>0.931924</td>\n",
       "      <td>0.938655</td>\n",
       "      <td>0.933259</td>\n",
       "      <td>0.939492</td>\n",
       "      <td>0.939294</td>\n",
       "      <td>0.943613</td>\n",
       "      <td>0.944489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.944497</td>\n",
       "      <td>0.944488</td>\n",
       "      <td>0.927921</td>\n",
       "      <td>0.925039</td>\n",
       "      <td>0.934119</td>\n",
       "      <td>0.932383</td>\n",
       "      <td>0.936000</td>\n",
       "      <td>0.931388</td>\n",
       "      <td>0.935250</td>\n",
       "      <td>0.940214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000158</th>\n",
       "      <td>0.948738</td>\n",
       "      <td>0.950139</td>\n",
       "      <td>0.937423</td>\n",
       "      <td>0.940137</td>\n",
       "      <td>0.933915</td>\n",
       "      <td>0.945687</td>\n",
       "      <td>0.942672</td>\n",
       "      <td>0.938922</td>\n",
       "      <td>0.944867</td>\n",
       "      <td>0.946364</td>\n",
       "      <td>...</td>\n",
       "      <td>0.936448</td>\n",
       "      <td>0.955136</td>\n",
       "      <td>0.930945</td>\n",
       "      <td>0.940298</td>\n",
       "      <td>0.947259</td>\n",
       "      <td>0.933552</td>\n",
       "      <td>0.941778</td>\n",
       "      <td>0.935631</td>\n",
       "      <td>0.937695</td>\n",
       "      <td>0.940228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM6569559  GSM6569560  GSM6569561  GSM6569562  GSM6569563  \\\n",
       "ID                                                                       \n",
       "cg00000029    0.296613    0.262710    0.328604    0.244043    0.305002   \n",
       "cg00000103    0.494243    0.507445    0.559102    0.496870    0.617299   \n",
       "cg00000109    0.864064    0.816083    0.846933    0.821823    0.838450   \n",
       "cg00000155    0.944174    0.944075    0.930635    0.931924    0.938655   \n",
       "cg00000158    0.948738    0.950139    0.937423    0.940137    0.933915   \n",
       "\n",
       "            GSM6569564  GSM6569565  GSM6569566  GSM6569567  GSM6569568  ...  \\\n",
       "ID                                                                      ...   \n",
       "cg00000029    0.351313    0.301256    0.223671    0.366115    0.352855  ...   \n",
       "cg00000103    0.567107    0.678045    0.555238    0.560232    0.524053  ...   \n",
       "cg00000109    0.825361    0.885593    0.862300    0.820759    0.851260  ...   \n",
       "cg00000155    0.933259    0.939492    0.939294    0.943613    0.944489  ...   \n",
       "cg00000158    0.945687    0.942672    0.938922    0.944867    0.946364  ...   \n",
       "\n",
       "            GSM6569581  GSM6569582  GSM6569583  GSM6569584  GSM6569585  \\\n",
       "ID                                                                       \n",
       "cg00000029    0.371031    0.308602    0.329137    0.309973    0.273167   \n",
       "cg00000103    0.621956    0.543419    0.558046    0.616681    0.618169   \n",
       "cg00000109    0.832698    0.831781    0.789656    0.817778    0.805174   \n",
       "cg00000155    0.944497    0.944488    0.927921    0.925039    0.934119   \n",
       "cg00000158    0.936448    0.955136    0.930945    0.940298    0.947259   \n",
       "\n",
       "            GSM6569586  GSM6569587  GSM6569588  GSM6569589  GSM6569590  \n",
       "ID                                                                      \n",
       "cg00000029    0.377755    0.266383    0.322724    0.287259    0.395006  \n",
       "cg00000103    0.586974    0.568257    0.627627    0.562612    0.662220  \n",
       "cg00000109    0.841091    0.806922    0.854048    0.799516    0.810282  \n",
       "cg00000155    0.932383    0.936000    0.931388    0.935250    0.940214  \n",
       "cg00000158    0.933552    0.941778    0.935631    0.937695    0.940228  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:14.107050774Z",
     "start_time": "2024-01-10T21:40:14.090977173Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cg00000029',\n",
       " 'cg00000103',\n",
       " 'cg00000109',\n",
       " 'cg00000155',\n",
       " 'cg00000158',\n",
       " 'cg00000165',\n",
       " 'cg00000221',\n",
       " 'cg00000236',\n",
       " 'cg00000289',\n",
       " 'cg00000292',\n",
       " 'cg00000321',\n",
       " 'cg00000363',\n",
       " 'cg00000540',\n",
       " 'cg00000579',\n",
       " 'cg00000596',\n",
       " 'cg00000622',\n",
       " 'cg00000658',\n",
       " 'cg00000714',\n",
       " 'cg00000721',\n",
       " 'cg00000734']"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_row_ids = genetic_data.index[:20].tolist()\n",
    "gene_row_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:16.445692615Z",
     "start_time": "2024-01-10T21:40:16.435117932Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['cg00000029', 'cg00000103', 'cg00000109', 'cg00000155', 'cg00000158', 'cg00000165', 'cg00000221', 'cg00000236', 'cg00000289', 'cg00000292', 'cg00000321', 'cg00000363', 'cg00000540', 'cg00000579', 'cg00000596', 'cg00000622', 'cg00000658', 'cg00000714', 'cg00000721', 'cg00000734']\\n\""
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n",
    "requires_gene_mapping = (True or False)\n",
    "\n",
    "Row headers:\n",
    "{gene_row_ids}\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "If not required, jump directly to the gene normalization step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.098774002Z",
     "start_time": "2024-01-10T21:40:17.094437169Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "requires_gene_mapping = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.399324928Z",
     "start_time": "2024-01-10T21:40:17.381475568Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978931927Z",
     "start_time": "2023-12-31T03:58:04.966328339Z"
    },
    "collapsed": false
   },
   "source": [
    "Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:17.944343822Z",
     "start_time": "2024-01-10T21:40:17.933815553Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    print(f'''\n",
    "    As a biomedical research team, we are analyzing a gene expression dataset, and find that its row headers are some identifiers related to genes:\n",
    "    {gene_row_ids}\n",
    "    To get the mapping from those identifiers to actual gene symbols, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the identifiers, and which key stores the gene symbols. Please strictly follow this format in your answer:\n",
    "    identifier_key = 'key_name1'\n",
    "    gene_symbol_key = 'key_name2'\n",
    "\n",
    "    Gene annotation dictionary:\n",
    "    {gene_annotation_summary}\n",
    "    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:18.245811669Z",
     "start_time": "2024-01-10T21:40:18.234480277Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'UCSC_RefGene_Name'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.559860684Z",
     "start_time": "2024-01-10T21:40:18.521711414Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1000 input query terms found no hit:\t['cg00000029', 'cg00000103', 'cg00000109', 'cg00000155', 'cg00000158', 'cg00000165', 'cg00000221', '\n",
      "1000 input query terms found no hit:\t['cg00026239', 'cg00026290', 'cg00026302', 'cg00026327', 'cg00026341', 'cg00026346', 'cg00026362', '\n",
      "1000 input query terms found no hit:\t['cg00053586', 'cg00053617', 'cg00053618', 'cg00053647', 'cg00053661', 'cg00053716', 'cg00053728', '\n",
      "1000 input query terms found no hit:\t['cg00081381', 'cg00081404', 'cg00081417', 'cg00081425', 'cg00081439', 'cg00081445', 'cg00081572', '\n",
      "1000 input query terms found no hit:\t['cg00107819', 'cg00107890', 'cg00107904', 'cg00107914', 'cg00107916', 'cg00107935', 'cg00107970', '\n",
      "1000 input query terms found no hit:\t['cg00136173', 'cg00136184', 'cg00136192', 'cg00136197', 'cg00136212', 'cg00136225', 'cg00136279', '\n",
      "1000 input query terms found no hit:\t['cg00164183', 'cg00164189', 'cg00164196', 'cg00164217', 'cg00164235', 'cg00164282', 'cg00164301', '\n",
      "1000 input query terms found no hit:\t['cg00193281', 'cg00193283', 'cg00193438', 'cg00193490', 'cg00193521', 'cg00193524', 'cg00193551', '\n",
      "1000 input query terms found no hit:\t['cg00223900', 'cg00223950', 'cg00223952', 'cg00223958', 'cg00223963', 'cg00224026', 'cg00224028', '\n",
      "1000 input query terms found no hit:\t['cg00252442', 'cg00252455', 'cg00252472', 'cg00252512', 'cg00252529', 'cg00252533', 'cg00252534', '\n",
      "1000 input query terms found no hit:\t['cg00282249', 'cg00282251', 'cg00282267', 'cg00282347', 'cg00282362', 'cg00282370', 'cg00282411', '\n",
      "1000 input query terms found no hit:\t['cg00312217', 'cg00312260', 'cg00312344', 'cg00312474', 'cg00312486', 'cg00312508', 'cg00312532', '\n",
      "1000 input query terms found no hit:\t['cg00339416', 'cg00339424', 'cg00339430', 'cg00339441', 'cg00339488', 'cg00339514', 'cg00339556', '\n",
      "1000 input query terms found no hit:\t['cg00366934', 'cg00366997', 'cg00367047', 'cg00367070', 'cg00367135', 'cg00367281', 'cg00367297', '\n",
      "1000 input query terms found no hit:\t['cg00394763', 'cg00394791', 'cg00394793', 'cg00394823', 'cg00394830', 'cg00394844', 'cg00394870', '\n",
      "1000 input query terms found no hit:\t['cg00421612', 'cg00421624', 'cg00421645', 'cg00421664', 'cg00421693', 'cg00421735', 'cg00421765', '\n",
      "1000 input query terms found no hit:\t['cg00451486', 'cg00451491', 'cg00451501', 'cg00451610', 'cg00451635', 'cg00451642', 'cg00451649', '\n",
      "1000 input query terms found no hit:\t['cg00477882', 'cg00477904', 'cg00477971', 'cg00477972', 'cg00477978', 'cg00477990', 'cg00478022', '\n",
      "1000 input query terms found no hit:\t['cg00505174', 'cg00505198', 'cg00505277', 'cg00505280', 'cg00505307', 'cg00505318', 'cg00505327', '\n",
      "1000 input query terms found no hit:\t['cg00532667', 'cg00532678', 'cg00532706', 'cg00532778', 'cg00532797', 'cg00532802', 'cg00532824', '\n",
      "1000 input query terms found no hit:\t['cg00560671', 'cg00560677', 'cg00560689', 'cg00560705', 'cg00560722', 'cg00560724', 'cg00560747', '\n",
      "1000 input query terms found no hit:\t['cg00587472', 'cg00587485', 'cg00587523', 'cg00587532', 'cg00587595', 'cg00587613', 'cg00587628', '\n",
      "1000 input query terms found no hit:\t['cg00614832', 'cg00614834', 'cg00614873', 'cg00614901', 'cg00614938', 'cg00614959', 'cg00614969', '\n",
      "1000 input query terms found no hit:\t['cg00642803', 'cg00642942', 'cg00642970', 'cg00643007', 'cg00643020', 'cg00643051', 'cg00643076', '\n",
      "1000 input query terms found no hit:\t['cg00672770', 'cg00672792', 'cg00672802', 'cg00672871', 'cg00672930', 'cg00672936', 'cg00672952', '\n",
      "1000 input query terms found no hit:\t['cg00700466', 'cg00700483', 'cg00700485', 'cg00700487', 'cg00700536', 'cg00700540', 'cg00700596', '\n",
      "1000 input query terms found no hit:\t['cg00729056', 'cg00729060', 'cg00729103', 'cg00729133', 'cg00729136', 'cg00729171', 'cg00729184', '\n",
      "1000 input query terms found no hit:\t['cg00756648', 'cg00756695', 'cg00756734', 'cg00756748', 'cg00756808', 'cg00756845', 'cg00756852', '\n",
      "1000 input query terms found no hit:\t['cg00785903', 'cg00785941', 'cg00785978', 'cg00785980', 'cg00786012', 'cg00786018', 'cg00786083', '\n",
      "1000 input query terms found no hit:\t['cg00814393', 'cg00814401', 'cg00814443', 'cg00814476', 'cg00814542', 'cg00814567', 'cg00814573', '\n",
      "1000 input query terms found no hit:\t['cg00844155', 'cg00844192', 'cg00844197', 'cg00844202', 'cg00844231', 'cg00844246', 'cg00844277', '\n",
      "1000 input query terms found no hit:\t['cg00873533', 'cg00873601', 'cg00873616', 'cg00873644', 'cg00873648', 'cg00873659', 'cg00873678', '\n",
      "1000 input query terms found no hit:\t['cg00903186', 'cg00903242', 'cg00903245', 'cg00903250', 'cg00903308', 'cg00903318', 'cg00903371', '\n",
      "1000 input query terms found no hit:\t['cg00931336', 'cg00931339', 'cg00931375', 'cg00931409', 'cg00931418', 'cg00931491', 'cg00931497', '\n",
      "1000 input query terms found no hit:\t['cg00960209', 'cg00960240', 'cg00960260', 'cg00960263', 'cg00960300', 'cg00960305', 'cg00960308', '\n",
      "1000 input query terms found no hit:\t['cg00987890', 'cg00987898', 'cg00987918', 'cg00987937', 'cg00988037', 'cg00988050', 'cg00988069', '\n",
      "1000 input query terms found no hit:\t['cg01016090', 'cg01016092', 'cg01016102', 'cg01016119', 'cg01016122', 'cg01016169', 'cg01016191', '\n",
      "1000 input query terms found no hit:\t['cg01044002', 'cg01044025', 'cg01044046', 'cg01044050', 'cg01044076', 'cg01044094', 'cg01044097', '\n",
      "1000 input query terms found no hit:\t['cg01072256', 'cg01072259', 'cg01072388', 'cg01072456', 'cg01072498', 'cg01072529', 'cg01072550', '\n",
      "1000 input query terms found no hit:\t['cg01101127', 'cg01101178', 'cg01101186', 'cg01101221', 'cg01101224', 'cg01101326', 'cg01101332', '\n",
      "1000 input query terms found no hit:\t['cg01129115', 'cg01129128', 'cg01129155', 'cg01129211', 'cg01129238', 'cg01129246', 'cg01129247', '\n",
      "1000 input query terms found no hit:\t['cg01157152', 'cg01157169', 'cg01157203', 'cg01157261', 'cg01157280', 'cg01157304', 'cg01157320', '\n",
      "1000 input query terms found no hit:\t['cg01185640', 'cg01185682', 'cg01185754', 'cg01185755', 'cg01185766', 'cg01185786', 'cg01185801', '\n",
      "1000 input query terms found no hit:\t['cg01216407', 'cg01216412', 'cg01216415', 'cg01216418', 'cg01216424', 'cg01216438', 'cg01216467', '\n",
      "1000 input query terms found no hit:\t['cg01245965', 'cg01245966', 'cg01245988', 'cg01246105', 'cg01246178', 'cg01246254', 'cg01246266', '\n",
      "1000 input query terms found no hit:\t['cg01276169', 'cg01276201', 'cg01276222', 'cg01276313', 'cg01276321', 'cg01276323', 'cg01276327', '\n",
      "1000 input query terms found no hit:\t['cg01305291', 'cg01305327', 'cg01305328', 'cg01305421', 'cg01305427', 'cg01305436', 'cg01305441', '\n",
      "1000 input query terms found no hit:\t['cg01334432', 'cg01334437', 'cg01334490', 'cg01334504', 'cg01334522', 'cg01334549', 'cg01334597', '\n",
      "1000 input query terms found no hit:\t['cg01363534', 'cg01363565', 'cg01363569', 'cg01363572', 'cg01363574', 'cg01363603', 'cg01363618', '\n",
      "1000 input query terms found no hit:\t['cg01392941', 'cg01393060', 'cg01393115', 'cg01393161', 'cg01393184', 'cg01393207', 'cg01393233', '\n",
      "1000 input query terms found no hit:\t['cg01421252', 'cg01421256', 'cg01421264', 'cg01421267', 'cg01421309', 'cg01421311', 'cg01421380', '\n",
      "1000 input query terms found no hit:\t['cg01449218', 'cg01449224', 'cg01449266', 'cg01449302', 'cg01449310', 'cg01449321', 'cg01449341', '\n",
      "1000 input query terms found no hit:\t['cg01480393', 'cg01480447', 'cg01480545', 'cg01480550', 'cg01480609', 'cg01480638', 'cg01480708', '\n",
      "1000 input query terms found no hit:\t['cg01510688', 'cg01510696', 'cg01510718', 'cg01510813', 'cg01510839', 'cg01510903', 'cg01510922', '\n",
      "1000 input query terms found no hit:\t['cg01538982', 'cg01538984', 'cg01538998', 'cg01539006', 'cg01539014', 'cg01539031', 'cg01539036', '\n",
      "1000 input query terms found no hit:\t['cg01567000', 'cg01567034', 'cg01567051', 'cg01567084', 'cg01567115', 'cg01567173', 'cg01567239', '\n",
      "1000 input query terms found no hit:\t['cg01596349', 'cg01596395', 'cg01596495', 'cg01596520', 'cg01596546', 'cg01596554', 'cg01596642', '\n",
      "1000 input query terms found no hit:\t['cg01625750', 'cg01625761', 'cg01625849', 'cg01625851', 'cg01625887', 'cg01625945', 'cg01625998', '\n",
      "1000 input query terms found no hit:\t['cg01656421', 'cg01656438', 'cg01656470', 'cg01656495', 'cg01656547', 'cg01656567', 'cg01656579', '\n",
      "1000 input query terms found no hit:\t['cg01690119', 'cg01690165', 'cg01690182', 'cg01690274', 'cg01690287', 'cg01690366', 'cg01690390', '\n",
      "1000 input query terms found no hit:\t['cg01719663', 'cg01719701', 'cg01719703', 'cg01719706', 'cg01719718', 'cg01719793', 'cg01719822', '\n",
      "1000 input query terms found no hit:\t['cg01750729', 'cg01750764', 'cg01750780', 'cg01750783', 'cg01750847', 'cg01750885', 'cg01750905', '\n",
      "1000 input query terms found no hit:\t['cg01780009', 'cg01780061', 'cg01780092', 'cg01780104', 'cg01780107', 'cg01780109', 'cg01780164', '\n",
      "1000 input query terms found no hit:\t['cg01807825', 'cg01807862', 'cg01807886', 'cg01807909', 'cg01807919', 'cg01807921', 'cg01807946', '\n",
      "1000 input query terms found no hit:\t['cg01836011', 'cg01836018', 'cg01836044', 'cg01836056', 'cg01836096', 'cg01836116', 'cg01836137', '\n",
      "1000 input query terms found no hit:\t['cg01867433', 'cg01867521', 'cg01867538', 'cg01867561', 'cg01867602', 'cg01867632', 'cg01867647', '\n",
      "1000 input query terms found no hit:\t['cg01895164', 'cg01895189', 'cg01895200', 'cg01895226', 'cg01895239', 'cg01895259', 'cg01895333', '\n",
      "1000 input query terms found no hit:\t['cg01923816', 'cg01923854', 'cg01923861', 'cg01923873', 'cg01923875', 'cg01923881', 'cg01923891', '\n",
      "1000 input query terms found no hit:\t['cg01952012', 'cg01952027', 'cg01952057', 'cg01952073', 'cg01952084', 'cg01952088', 'cg01952185', '\n",
      "1000 input query terms found no hit:\t['cg01980115', 'cg01980126', 'cg01980174', 'cg01980176', 'cg01980222', 'cg01980225', 'cg01980298', '\n",
      "1000 input query terms found no hit:\t['cg02009879', 'cg02009891', 'cg02009895', 'cg02009964', 'cg02009986', 'cg02010004', 'cg02010013', '\n",
      "1000 input query terms found no hit:\t['cg02038437', 'cg02038468', 'cg02038492', 'cg02038515', 'cg02038558', 'cg02038561', 'cg02038651', '\n",
      "1000 input query terms found no hit:\t['cg02066994', 'cg02067022', 'cg02067029', 'cg02067030', 'cg02067098', 'cg02067137', 'cg02067211', '\n",
      "1000 input query terms found no hit:\t['cg02096303', 'cg02096324', 'cg02096368', 'cg02096384', 'cg02096396', 'cg02096397', 'cg02096431', '\n",
      "1000 input query terms found no hit:\t['cg02125654', 'cg02125702', 'cg02125727', 'cg02125896', 'cg02125910', 'cg02125938', 'cg02125941', '\n",
      "1000 input query terms found no hit:\t['cg02156567', 'cg02156570', 'cg02156591', 'cg02156598', 'cg02156680', 'cg02156692', 'cg02156704', '\n",
      "1000 input query terms found no hit:\t['cg02186610', 'cg02186614', 'cg02186707', 'cg02186735', 'cg02186748', 'cg02186756', 'cg02186769', '\n",
      "1000 input query terms found no hit:\t['cg02217859', 'cg02217895', 'cg02217955', 'cg02217975', 'cg02217979', 'cg02218000', 'cg02218090', '\n",
      "1000 input query terms found no hit:\t['cg02247163', 'cg02247175', 'cg02247178', 'cg02247183', 'cg02247211', 'cg02247227', 'cg02247267', '\n",
      "1000 input query terms found no hit:\t['cg02276314', 'cg02276361', 'cg02276490', 'cg02276537', 'cg02276600', 'cg02276614', 'cg02276665', '\n",
      "1000 input query terms found no hit:\t['cg02303592', 'cg02303647', 'cg02303651', 'cg02303677', 'cg02303701', 'cg02303726', 'cg02303737', '\n",
      "1000 input query terms found no hit:\t['cg02331450', 'cg02331522', 'cg02331561', 'cg02331583', 'cg02331594', 'cg02331649', 'cg02331655', '\n",
      "1000 input query terms found no hit:\t['cg02361027', 'cg02361121', 'cg02361126', 'cg02361130', 'cg02361196', 'cg02361229', 'cg02361260', '\n",
      "1000 input query terms found no hit:\t['cg02389513', 'cg02389555', 'cg02389566', 'cg02389569', 'cg02389590', 'cg02389617', 'cg02389631', '\n",
      "1000 input query terms found no hit:\t['cg02417838', 'cg02417857', 'cg02417860', 'cg02417862', 'cg02417875', 'cg02418025', 'cg02418055', '\n",
      "1000 input query terms found no hit:\t['cg02451467', 'cg02451499', 'cg02451502', 'cg02451508', 'cg02451516', 'cg02451536', 'cg02451548', '\n",
      "1000 input query terms found no hit:\t['cg02479692', 'cg02479744', 'cg02479755', 'cg02479773', 'cg02479782', 'cg02479785', 'cg02479789', '\n",
      "1000 input query terms found no hit:\t['cg02507604', 'cg02507657', 'cg02507663', 'cg02507665', 'cg02507677', 'cg02507726', 'cg02507810', '\n",
      "1000 input query terms found no hit:\t['cg02536541', 'cg02536545', 'cg02536588', 'cg02536620', 'cg02536625', 'cg02536637', 'cg02536683', '\n",
      "1000 input query terms found no hit:\t['cg02565647', 'cg02565650', 'cg02565669', 'cg02565672', 'cg02565702', 'cg02565715', 'cg02565730', '\n",
      "1000 input query terms found no hit:\t['cg02594500', 'cg02594532', 'cg02594607', 'cg02594615', 'cg02594677', 'cg02594687', 'cg02594804', '\n",
      "1000 input query terms found no hit:\t['cg02623114', 'cg02623175', 'cg02623200', 'cg02623234', 'cg02623256', 'cg02623259', 'cg02623354', '\n",
      "1000 input query terms found no hit:\t['cg02648692', 'cg02648724', 'cg02648738', 'cg02648746', 'cg02648768', 'cg02648821', 'cg02648847', '\n",
      "1000 input query terms found no hit:\t['cg02679208', 'cg02679214', 'cg02679238', 'cg02679273', 'cg02679291', 'cg02679294', 'cg02679322', '\n",
      "1000 input query terms found no hit:\t['cg02709807', 'cg02709812', 'cg02709834', 'cg02709840', 'cg02709877', 'cg02709907', 'cg02709924', '\n",
      "1000 input query terms found no hit:\t['cg02738772', 'cg02738868', 'cg02738938', 'cg02738977', 'cg02738983', 'cg02738988', 'cg02739013', '\n",
      "1000 input query terms found no hit:\t['cg02767665', 'cg02767668', 'cg02767673', 'cg02767759', 'cg02767771', 'cg02767777', 'cg02767788', '\n",
      "1000 input query terms found no hit:\t['cg02797166', 'cg02797183', 'cg02797195', 'cg02797201', 'cg02797217', 'cg02797233', 'cg02797271', '\n",
      "1000 input query terms found no hit:\t['cg02827328', 'cg02827336', 'cg02827340', 'cg02827371', 'cg02827386', 'cg02827457', 'cg02827459', '\n",
      "1000 input query terms found no hit:\t['cg02856379', 'cg02856402', 'cg02856420', 'cg02856462', 'cg02856473', 'cg02856475', 'cg02856481', '\n",
      "1000 input query terms found no hit:\t['cg02885095', 'cg02885120', 'cg02885171', 'cg02885188', 'cg02885259', 'cg02885428', 'cg02885454', '\n",
      "1000 input query terms found no hit:\t['cg02912913', 'cg02912914', 'cg02912937', 'cg02912949', 'cg02912956', 'cg02912973', 'cg02913035', '\n",
      "1000 input query terms found no hit:\t['cg02940523', 'cg02940536', 'cg02940555', 'cg02940562', 'cg02940599', 'cg02940615', 'cg02940645', '\n",
      "1000 input query terms found no hit:\t['cg02968658', 'cg02968715', 'cg02968723', 'cg02968741', 'cg02968751', 'cg02968760', 'cg02968765', '\n",
      "1000 input query terms found no hit:\t['cg02996583', 'cg02996602', 'cg02996608', 'cg02996630', 'cg02996639', 'cg02996672', 'cg02996684', '\n",
      "1000 input query terms found no hit:\t['cg03026462', 'cg03026483', 'cg03026524', 'cg03026574', 'cg03026611', 'cg03026689', 'cg03026698', '\n",
      "1000 input query terms found no hit:\t['cg03055065', 'cg03055131', 'cg03055158', 'cg03055238', 'cg03055241', 'cg03055271', 'cg03055279', '\n",
      "1000 input query terms found no hit:\t['cg03085262', 'cg03085275', 'cg03085299', 'cg03085312', 'cg03085335', 'cg03085364', 'cg03085377', '\n",
      "1000 input query terms found no hit:\t['cg03116016', 'cg03116030', 'cg03116035', 'cg03116124', 'cg03116148', 'cg03116171', 'cg03116182', '\n",
      "1000 input query terms found no hit:\t['cg03146065', 'cg03146073', 'cg03146079', 'cg03146103', 'cg03146104', 'cg03146143', 'cg03146154', '\n",
      "1000 input query terms found no hit:\t['cg03176205', 'cg03176282', 'cg03176298', 'cg03176305', 'cg03176386', 'cg03176411', 'cg03176423', '\n",
      "1000 input query terms found no hit:\t['cg03205581', 'cg03205584', 'cg03205601', 'cg03205750', 'cg03205764', 'cg03205844', 'cg03205986', '\n",
      "1000 input query terms found no hit:\t['cg03233793', 'cg03233876', 'cg03233885', 'cg03233896', 'cg03233918', 'cg03233931', 'cg03233940', '\n",
      "1000 input query terms found no hit:\t['cg03264444', 'cg03264464', 'cg03264503', 'cg03264526', 'cg03264550', 'cg03264585', 'cg03264601', '\n",
      "1000 input query terms found no hit:\t['cg03294741', 'cg03294833', 'cg03294907', 'cg03294925', 'cg03294961', 'cg03295019', 'cg03295070', '\n",
      "1000 input query terms found no hit:\t['cg03323794', 'cg03323816', 'cg03323895', 'cg03323911', 'cg03323924', 'cg03323938', 'cg03323946', '\n",
      "1000 input query terms found no hit:\t['cg03352242', 'cg03352287', 'cg03352322', 'cg03352332', 'cg03352346', 'cg03352347', 'cg03352396', '\n",
      "1000 input query terms found no hit:\t['cg03384175', 'cg03384294', 'cg03384304', 'cg03384391', 'cg03384413', 'cg03384419', 'cg03384426', '\n",
      "1000 input query terms found no hit:\t['cg03413100', 'cg03413134', 'cg03413144', 'cg03413202', 'cg03413241', 'cg03413251', 'cg03413253', '\n",
      "1000 input query terms found no hit:\t['cg03442477', 'cg03442510', 'cg03442523', 'cg03442534', 'cg03442540', 'cg03442545', 'cg03442564', '\n",
      "1000 input query terms found no hit:\t['cg03470745', 'cg03470754', 'cg03470772', 'cg03470780', 'cg03470837', 'cg03470841', 'cg03470856', '\n",
      "1000 input query terms found no hit:\t['cg03502630', 'cg03502656', 'cg03502757', 'cg03502767', 'cg03502791', 'cg03502812', 'cg03502863', '\n",
      "1000 input query terms found no hit:\t['cg03530994', 'cg03531013', 'cg03531100', 'cg03531140', 'cg03531177', 'cg03531194', 'cg03531211', '\n",
      "1000 input query terms found no hit:\t['cg03558233', 'cg03558255', 'cg03558269', 'cg03558299', 'cg03558310', 'cg03558326', 'cg03558338', '\n",
      "1000 input query terms found no hit:\t['cg03587843', 'cg03587877', 'cg03587884', 'cg03587887', 'cg03587907', 'cg03587917', 'cg03587925', '\n",
      "1000 input query terms found no hit:\t['cg03617023', 'cg03617037', 'cg03617054', 'cg03617106', 'cg03617187', 'cg03617221', 'cg03617314', '\n",
      "1000 input query terms found no hit:\t['cg03646393', 'cg03646441', 'cg03646500', 'cg03646501', 'cg03646542', 'cg03646557', 'cg03646564', '\n",
      "1000 input query terms found no hit:\t['cg03674826', 'cg03674841', 'cg03674851', 'cg03674870', 'cg03674955', 'cg03674983', 'cg03675047', '\n",
      "1000 input query terms found no hit:\t['cg03705718', 'cg03705727', 'cg03705758', 'cg03705784', 'cg03705817', 'cg03705827', 'cg03705875', '\n",
      "1000 input query terms found no hit:\t['cg03734391', 'cg03734392', 'cg03734401', 'cg03734402', 'cg03734437', 'cg03734501', 'cg03734533', '\n",
      "1000 input query terms found no hit:\t['cg03763391', 'cg03763444', 'cg03763452', 'cg03763457', 'cg03763508', 'cg03763518', 'cg03763570', '\n",
      "1000 input query terms found no hit:\t['cg03794124', 'cg03794154', 'cg03794166', 'cg03794213', 'cg03794214', 'cg03794218', 'cg03794219', '\n",
      "1000 input query terms found no hit:\t['cg03825574', 'cg03825607', 'cg03825670', 'cg03825675', 'cg03825683', 'cg03825736', 'cg03825764', '\n",
      "1000 input query terms found no hit:\t['cg03856026', 'cg03856044', 'cg03856079', 'cg03856109', 'cg03856126', 'cg03856131', 'cg03856144', '\n",
      "1000 input query terms found no hit:\t['cg03886909', 'cg03886983', 'cg03886989', 'cg03887008', 'cg03887049', 'cg03887092', 'cg03887094', '\n",
      "1000 input query terms found no hit:\t['cg03916832', 'cg03916864', 'cg03916902', 'cg03916903', 'cg03916908', 'cg03916909', 'cg03916912', '\n",
      "1000 input query terms found no hit:\t['cg03948013', 'cg03948025', 'cg03948048', 'cg03948067', 'cg03948181', 'cg03948182', 'cg03948194', '\n",
      "1000 input query terms found no hit:\t['cg03977449', 'cg03977478', 'cg03977586', 'cg03977592', 'cg03977651', 'cg03977657', 'cg03977696', '\n",
      "1000 input query terms found no hit:\t['cg04006718', 'cg04006722', 'cg04006766', 'cg04006839', 'cg04006871', 'cg04006891', 'cg04006913', '\n",
      "1000 input query terms found no hit:\t['cg04037259', 'cg04037327', 'cg04037369', 'cg04037372', 'cg04037398', 'cg04037410', 'cg04037447', '\n",
      "1000 input query terms found no hit:\t['cg04068910', 'cg04068914', 'cg04068959', 'cg04068966', 'cg04068975', 'cg04068980', 'cg04068989', '\n",
      "1000 input query terms found no hit:\t['cg04098270', 'cg04098280', 'cg04098288', 'cg04098293', 'cg04098297', 'cg04098302', 'cg04098314', '\n",
      "1000 input query terms found no hit:\t['cg04124888', 'cg04124909', 'cg04124916', 'cg04124962', 'cg04124966', 'cg04124981', 'cg04124982', '\n",
      "1000 input query terms found no hit:\t['cg04153092', 'cg04153116', 'cg04153130', 'cg04153135', 'cg04153152', 'cg04153193', 'cg04153241', '\n",
      "1000 input query terms found no hit:\t['cg04182040', 'cg04182076', 'cg04182114', 'cg04182126', 'cg04182170', 'cg04182197', 'cg04182207', '\n",
      "1000 input query terms found no hit:\t['cg04212812', 'cg04212815', 'cg04212842', 'cg04212846', 'cg04212884', 'cg04212926', 'cg04212933', '\n",
      "1000 input query terms found no hit:\t['cg04240493', 'cg04240504', 'cg04240525', 'cg04240548', 'cg04240553', 'cg04240583', 'cg04240588', '\n",
      "1000 input query terms found no hit:\t['cg04269700', 'cg04269747', 'cg04269762', 'cg04269791', 'cg04269848', 'cg04269878', 'cg04269879', '\n",
      "1000 input query terms found no hit:\t['cg04299757', 'cg04299787', 'cg04299808', 'cg04299914', 'cg04299936', 'cg04300014', 'cg04300065', '\n",
      "1000 input query terms found no hit:\t['cg04329560', 'cg04329613', 'cg04329659', 'cg04329737', 'cg04329845', 'cg04329870', 'cg04329876', '\n",
      "1000 input query terms found no hit:\t['cg04359250', 'cg04359252', 'cg04359295', 'cg04359302', 'cg04359324', 'cg04359351', 'cg04359365', '\n",
      "1000 input query terms found no hit:\t['cg04390652', 'cg04390655', 'cg04390667', 'cg04390683', 'cg04390734', 'cg04390829', 'cg04390831', '\n",
      "1000 input query terms found no hit:\t['cg04420141', 'cg04420196', 'cg04420217', 'cg04420253', 'cg04420275', 'cg04420294', 'cg04420309', '\n",
      "1000 input query terms found no hit:\t['cg04451622', 'cg04451652', 'cg04451691', 'cg04451699', 'cg04451703', 'cg04451736', 'cg04451748', '\n",
      "1000 input query terms found no hit:\t['cg04482072', 'cg04482075', 'cg04482088', 'cg04482110', 'cg04482120', 'cg04482131', 'cg04482148', '\n",
      "1000 input query terms found no hit:\t['cg04511182', 'cg04511195', 'cg04511291', 'cg04511352', 'cg04511411', 'cg04511418', 'cg04511424', '\n",
      "1000 input query terms found no hit:\t['cg04541078', 'cg04541114', 'cg04541146', 'cg04541154', 'cg04541218', 'cg04541228', 'cg04541281', '\n",
      "1000 input query terms found no hit:\t['cg04569233', 'cg04569313', 'cg04569318', 'cg04569364', 'cg04569381', 'cg04569424', 'cg04569429', '\n",
      "1000 input query terms found no hit:\t['cg04597894', 'cg04597945', 'cg04597957', 'cg04597985', 'cg04597991', 'cg04598035', 'cg04598070', '\n",
      "1000 input query terms found no hit:\t['cg04628240', 'cg04628293', 'cg04628308', 'cg04628320', 'cg04628328', 'cg04628369', 'cg04628393', '\n",
      "1000 input query terms found no hit:\t['cg04661246', 'cg04661335', 'cg04661346', 'cg04661358', 'cg04661382', 'cg04661436', 'cg04661457', '\n",
      "1000 input query terms found no hit:\t['cg04690793', 'cg04690820', 'cg04690834', 'cg04690840', 'cg04690917', 'cg04690927', 'cg04690994', '\n",
      "1000 input query terms found no hit:\t['cg04722610', 'cg04722620', 'cg04722621', 'cg04722637', 'cg04722643', 'cg04722675', 'cg04722735', '\n",
      "1000 input query terms found no hit:\t['cg04750804', 'cg04750910', 'cg04750995', 'cg04751006', 'cg04751009', 'cg04751028', 'cg04751029', '\n",
      "1000 input query terms found no hit:\t['cg04781709', 'cg04781741', 'cg04781764', 'cg04781796', 'cg04781820', 'cg04781839', 'cg04781873', '\n",
      "1000 input query terms found no hit:\t['cg04810284', 'cg04810287', 'cg04810377', 'cg04810396', 'cg04810466', 'cg04810656', 'cg04810660', '\n",
      "1000 input query terms found no hit:\t['cg04840669', 'cg04840679', 'cg04840702', 'cg04840732', 'cg04840739', 'cg04840761', 'cg04840800', '\n",
      "1000 input query terms found no hit:\t['cg04869063', 'cg04869116', 'cg04869122', 'cg04869136', 'cg04869206', 'cg04869269', 'cg04869310', '\n",
      "1000 input query terms found no hit:\t['cg04897230', 'cg04897245', 'cg04897299', 'cg04897348', 'cg04897370', 'cg04897406', 'cg04897431', '\n",
      "1000 input query terms found no hit:\t['cg04924696', 'cg04924736', 'cg04924893', 'cg04924907', 'cg04924962', 'cg04925022', 'cg04925032', '\n",
      "1000 input query terms found no hit:\t['cg04954795', 'cg04954797', 'cg04954894', 'cg04954926', 'cg04954975', 'cg04954994', 'cg04955051', '\n",
      "1000 input query terms found no hit:\t['cg04985004', 'cg04985016', 'cg04985020', 'cg04985032', 'cg04985036', 'cg04985052', 'cg04985089', '\n",
      "1000 input query terms found no hit:\t['cg05013129', 'cg05013166', 'cg05013218', 'cg05013245', 'cg05013250', 'cg05013284', 'cg05013337', '\n",
      "1000 input query terms found no hit:\t['cg05041997', 'cg05042034', 'cg05042038', 'cg05042043', 'cg05042077', 'cg05042088', 'cg05042110', '\n",
      "1000 input query terms found no hit:\t['cg05069885', 'cg05069909', 'cg05069934', 'cg05069950', 'cg05069952', 'cg05069953', 'cg05070042', '\n",
      "1000 input query terms found no hit:\t['cg05099508', 'cg05099537', 'cg05099576', 'cg05099596', 'cg05099600', 'cg05099605', 'cg05099692', '\n",
      "1000 input query terms found no hit:\t['cg05131524', 'cg05131526', 'cg05131533', 'cg05131545', 'cg05131552', 'cg05131620', 'cg05131623', '\n",
      "1000 input query terms found no hit:\t['cg05160563', 'cg05160573', 'cg05160592', 'cg05160614', 'cg05160664', 'cg05160678', 'cg05160751', '\n",
      "1000 input query terms found no hit:\t['cg05190897', 'cg05190930', 'cg05190931', 'cg05190994', 'cg05191006', 'cg05191013', 'cg05191034', '\n",
      "1000 input query terms found no hit:\t['cg05219310', 'cg05219378', 'cg05219421', 'cg05219423', 'cg05219430', 'cg05219445', 'cg05219464', '\n",
      "1000 input query terms found no hit:\t['cg05247426', 'cg05247515', 'cg05247640', 'cg05247650', 'cg05247767', 'cg05247773', 'cg05247780', '\n",
      "1000 input query terms found no hit:\t['cg05277491', 'cg05277504', 'cg05277736', 'cg05277793', 'cg05277800', 'cg05277865', 'cg05277881', '\n",
      "1000 input query terms found no hit:\t['cg05306815', 'cg05306824', 'cg05306831', 'cg05306856', 'cg05306860', 'cg05306874', 'cg05306897', '\n",
      "1000 input query terms found no hit:\t['cg05337743', 'cg05337753', 'cg05337761', 'cg05337769', 'cg05337779', 'cg05337806', 'cg05337936', '\n",
      "1000 input query terms found no hit:\t['cg05364761', 'cg05364762', 'cg05364766', 'cg05364767', 'cg05364829', 'cg05364839', 'cg05364884', '\n",
      "1000 input query terms found no hit:\t['cg05392141', 'cg05392160', 'cg05392169', 'cg05392222', 'cg05392230', 'cg05392244', 'cg05392263', '\n",
      "1000 input query terms found no hit:\t['cg05422836', 'cg05422883', 'cg05422897', 'cg05422916', 'cg05422920', 'cg05422925', 'cg05422948', '\n",
      "1000 input query terms found no hit:\t['cg05456789', 'cg05456799', 'cg05456810', 'cg05456868', 'cg05456890', 'cg05456894', 'cg05456909', '\n",
      "1000 input query terms found no hit:\t['cg05487507', 'cg05487508', 'cg05487589', 'cg05487635', 'cg05487650', 'cg05487654', 'cg05487664', '\n",
      "1000 input query terms found no hit:\t['cg05517387', 'cg05517400', 'cg05517418', 'cg05517438', 'cg05517447', 'cg05517456', 'cg05517471', '\n",
      "1000 input query terms found no hit:\t['cg05549498', 'cg05549544', 'cg05549554', 'cg05549575', 'cg05549611', 'cg05549626', 'cg05549638', '\n",
      "1000 input query terms found no hit:\t['cg05581264', 'cg05581275', 'cg05581297', 'cg05581313', 'cg05581314', 'cg05581388', 'cg05581394', '\n",
      "1000 input query terms found no hit:\t['cg05611617', 'cg05611648', 'cg05611670', 'cg05611684', 'cg05611687', 'cg05611719', 'cg05611722', '\n",
      "1000 input query terms found no hit:\t['cg05641351', 'cg05641432', 'cg05641439', 'cg05641446', 'cg05641515', 'cg05641529', 'cg05641535', '\n",
      "1000 input query terms found no hit:\t['cg05670439', 'cg05670459', 'cg05670461', 'cg05670472', 'cg05670486', 'cg05670488', 'cg05670509', '\n",
      "1000 input query terms found no hit:\t['cg05702052', 'cg05702092', 'cg05702161', 'cg05702207', 'cg05702218', 'cg05702234', 'cg05702280', '\n",
      "1000 input query terms found no hit:\t['cg05731859', 'cg05731861', 'cg05731870', 'cg05731995', 'cg05732044', 'cg05732081', 'cg05732089', '\n",
      "1000 input query terms found no hit:\t['cg05764376', 'cg05764393', 'cg05764403', 'cg05764422', 'cg05764431', 'cg05764470', 'cg05764506', '\n",
      "1000 input query terms found no hit:\t['cg05793375', 'cg05793390', 'cg05793391', 'cg05793404', 'cg05793406', 'cg05793409', 'cg05793444', '\n",
      "1000 input query terms found no hit:\t['cg05825459', 'cg05825520', 'cg05825523', 'cg05825555', 'cg05825559', 'cg05825578', 'cg05825596', '\n",
      "1000 input query terms found no hit:\t['cg05857377', 'cg05857597', 'cg05857643', 'cg05857728', 'cg05857758', 'cg05857773', 'cg05857825', '\n",
      "1000 input query terms found no hit:\t['cg05885647', 'cg05885668', 'cg05885676', 'cg05885688', 'cg05885717', 'cg05885720', 'cg05885732', '\n",
      "1000 input query terms found no hit:\t['cg05913672', 'cg05913684', 'cg05913712', 'cg05913737', 'cg05913826', 'cg05913846', 'cg05913902', '\n",
      "1000 input query terms found no hit:\t['cg05941060', 'cg05941106', 'cg05941108', 'cg05941116', 'cg05941129', 'cg05941153', 'cg05941155', '\n",
      "1000 input query terms found no hit:\t['cg05970080', 'cg05970116', 'cg05970164', 'cg05970215', 'cg05970234', 'cg05970238', 'cg05970288', '\n",
      "1000 input query terms found no hit:\t['cg06000470', 'cg06000491', 'cg06000556', 'cg06000564', 'cg06000584', 'cg06000610', 'cg06000635', '\n",
      "1000 input query terms found no hit:\t['cg06031737', 'cg06031740', 'cg06031796', 'cg06031797', 'cg06031841', 'cg06031849', 'cg06031853', '\n",
      "1000 input query terms found no hit:\t['cg06062703', 'cg06062730', 'cg06062763', 'cg06062784', 'cg06062821', 'cg06062823', 'cg06062848', '\n",
      "1000 input query terms found no hit:\t['cg06092310', 'cg06092312', 'cg06092382', 'cg06092502', 'cg06092573', 'cg06092604', 'cg06092636', '\n",
      "1000 input query terms found no hit:\t['cg06122644', 'cg06122660', 'cg06122697', 'cg06122740', 'cg06122773', 'cg06122825', 'cg06122838', '\n",
      "1000 input query terms found no hit:\t['cg06154159', 'cg06154200', 'cg06154311', 'cg06154313', 'cg06154315', 'cg06154338', 'cg06154408', '\n",
      "1000 input query terms found no hit:\t['cg06185515', 'cg06185532', 'cg06185555', 'cg06185577', 'cg06185593', 'cg06185613', 'cg06185657', '\n",
      "1000 input query terms found no hit:\t['cg06215416', 'cg06215447', 'cg06215457', 'cg06215502', 'cg06215509', 'cg06215536', 'cg06215556', '\n",
      "1000 input query terms found no hit:\t['cg06246888', 'cg06246900', 'cg06246935', 'cg06246941', 'cg06246950', 'cg06246967', 'cg06246999', '\n",
      "1000 input query terms found no hit:\t['cg06280360', 'cg06280368', 'cg06280460', 'cg06280512', 'cg06280523', 'cg06280537', 'cg06280568', '\n",
      "1000 input query terms found no hit:\t['cg06312004', 'cg06312024', 'cg06312055', 'cg06312066', 'cg06312072', 'cg06312106', 'cg06312132', '\n",
      "1000 input query terms found no hit:\t['cg06343183', 'cg06343212', 'cg06343233', 'cg06343267', 'cg06343288', 'cg06343355', 'cg06343561', '\n",
      "1000 input query terms found no hit:\t['cg06374862', 'cg06374892', 'cg06374899', 'cg06374918', 'cg06374962', 'cg06375042', 'cg06375062', '\n",
      "1000 input query terms found no hit:\t['cg06405129', 'cg06405146', 'cg06405161', 'cg06405186', 'cg06405206', 'cg06405219', 'cg06405222', '\n",
      "1000 input query terms found no hit:\t['cg06437001', 'cg06437004', 'cg06437006', 'cg06437030', 'cg06437052', 'cg06437100', 'cg06437110', '\n",
      "1000 input query terms found no hit:\t['cg06467579', 'cg06467587', 'cg06467600', 'cg06467610', 'cg06467627', 'cg06467636', 'cg06467638', '\n",
      "1000 input query terms found no hit:\t['cg06499262', 'cg06499265', 'cg06499271', 'cg06499344', 'cg06499359', 'cg06499415', 'cg06499424', '\n",
      "1000 input query terms found no hit:\t['cg06529181', 'cg06529201', 'cg06529227', 'cg06529255', 'cg06529264', 'cg06529318', 'cg06529340', '\n",
      "1000 input query terms found no hit:\t['cg06558502', 'cg06558519', 'cg06558623', 'cg06558640', 'cg06558674', 'cg06558703', 'cg06558765', '\n",
      "1000 input query terms found no hit:\t['cg06587931', 'cg06587945', 'cg06587962', 'cg06587969', 'cg06587976', 'cg06588001', 'cg06588010', '\n",
      "1000 input query terms found no hit:\t['cg06618138', 'cg06618163', 'cg06618254', 'cg06618277', 'cg06618298', 'cg06618322', 'cg06618356', '\n",
      "1000 input query terms found no hit:\t['cg06646218', 'cg06646222', 'cg06646234', 'cg06646246', 'cg06646262', 'cg06646276', 'cg06646285', '\n",
      "1000 input query terms found no hit:\t['cg06675660', 'cg06675690', 'cg06675705', 'cg06675741', 'cg06675775', 'cg06675781', 'cg06675784', '\n",
      "1000 input query terms found no hit:\t['cg06704969', 'cg06704974', 'cg06705004', 'cg06705017', 'cg06705045', 'cg06705062', 'cg06705064', '\n",
      "1000 input query terms found no hit:\t['cg06732599', 'cg06732663', 'cg06732666', 'cg06732667', 'cg06732684', 'cg06732687', 'cg06732688', '\n",
      "1000 input query terms found no hit:\t['cg06759890', 'cg06759901', 'cg06759944', 'cg06759971', 'cg06759993', 'cg06760004', 'cg06760018', '\n",
      "1000 input query terms found no hit:\t['cg06789134', 'cg06789144', 'cg06789190', 'cg06789194', 'cg06789233', 'cg06789340', 'cg06789383', '\n",
      "1000 input query terms found no hit:\t['cg06817609', 'cg06817610', 'cg06817614', 'cg06817634', 'cg06817639', 'cg06817658', 'cg06817669', '\n",
      "1000 input query terms found no hit:\t['cg06844749', 'cg06844837', 'cg06844845', 'cg06844864', 'cg06844930', 'cg06844932', 'cg06844936', '\n",
      "1000 input query terms found no hit:\t['cg06872673', 'cg06872683', 'cg06872687', 'cg06872705', 'cg06872709', 'cg06872721', 'cg06872770', '\n",
      "1000 input query terms found no hit:\t['cg06899844', 'cg06899912', 'cg06899953', 'cg06899970', 'cg06899976', 'cg06899985', 'cg06900004', '\n",
      "1000 input query terms found no hit:\t['cg06929186', 'cg06929216', 'cg06929230', 'cg06929287', 'cg06929324', 'cg06929371', 'cg06929414', '\n",
      "1000 input query terms found no hit:\t['cg06958747', 'cg06958766', 'cg06958829', 'cg06958842', 'cg06958882', 'cg06958924', 'cg06958930', '\n",
      "1000 input query terms found no hit:\t['cg06988133', 'cg06988187', 'cg06988196', 'cg06988236', 'cg06988251', 'cg06988310', 'cg06988336', '\n",
      "1000 input query terms found no hit:\t['cg07017374', 'cg07017379', 'cg07017420', 'cg07017437', 'cg07017454', 'cg07017477', 'cg07017486', '\n",
      "1000 input query terms found no hit:\t['cg07048244', 'cg07048256', 'cg07048306', 'cg07048321', 'cg07048326', 'cg07048348', 'cg07048375', '\n",
      "1000 input query terms found no hit:\t['cg07077466', 'cg07077476', 'cg07077507', 'cg07077525', 'cg07077665', 'cg07077694', 'cg07077697', '\n",
      "1000 input query terms found no hit:\t['cg07106654', 'cg07106691', 'cg07106737', 'cg07106766', 'cg07106775', 'cg07106794', 'cg07106832', '\n",
      "1000 input query terms found no hit:\t['cg07135629', 'cg07135647', 'cg07135660', 'cg07135664', 'cg07135701', 'cg07135706', 'cg07135723', '\n",
      "1000 input query terms found no hit:\t['cg07162704', 'cg07162727', 'cg07162735', 'cg07162768', 'cg07162801', 'cg07162820', 'cg07162826', '\n",
      "1000 input query terms found no hit:\t['cg07191957', 'cg07192043', 'cg07192047', 'cg07192082', 'cg07192240', 'cg07192243', 'cg07192275', '\n",
      "1000 input query terms found no hit:\t['cg07221824', 'cg07221830', 'cg07221832', 'cg07221855', 'cg07221885', 'cg07221967', 'cg07222015', '\n",
      "1000 input query terms found no hit:\t['cg07251783', 'cg07251788', 'cg07251857', 'cg07251872', 'cg07251878', 'cg07251887', 'cg07251892', '\n",
      "1000 input query terms found no hit:\t['cg07281821', 'cg07281879', 'cg07281938', 'cg07281948', 'cg07281959', 'cg07282058', 'cg07282076', '\n",
      "1000 input query terms found no hit:\t['cg07313437', 'cg07313438', 'cg07313441', 'cg07313485', 'cg07313499', 'cg07313502', 'cg07313504', '\n",
      "1000 input query terms found no hit:\t['cg07342403', 'cg07342512', 'cg07342545', 'cg07342559', 'cg07342573', 'cg07342618', 'cg07342631', '\n",
      "1000 input query terms found no hit:\t['cg07372060', 'cg07372166', 'cg07372173', 'cg07372195', 'cg07372207', 'cg07372330', 'cg07372360', '\n",
      "1000 input query terms found no hit:\t['cg07400751', 'cg07400877', 'cg07401028', 'cg07401045', 'cg07401063', 'cg07401067', 'cg07401103', '\n",
      "1000 input query terms found no hit:\t['cg07434038', 'cg07434077', 'cg07434082', 'cg07434166', 'cg07434188', 'cg07434253', 'cg07434260', '\n",
      "1000 input query terms found no hit:\t['cg07464204', 'cg07464206', 'cg07464212', 'cg07464217', 'cg07464229', 'cg07464248', 'cg07464250', '\n",
      "1000 input query terms found no hit:\t['cg07492518', 'cg07492590', 'cg07492636', 'cg07492660', 'cg07492680', 'cg07492700', 'cg07492739', '\n",
      "1000 input query terms found no hit:\t['cg07522962', 'cg07522970', 'cg07522981', 'cg07523043', 'cg07523071', 'cg07523089', 'cg07523098', '\n",
      "1000 input query terms found no hit:\t['cg07553412', 'cg07553475', 'cg07553500', 'cg07553512', 'cg07553518', 'cg07553580', 'cg07553595', '\n",
      "1000 input query terms found no hit:\t['cg07584150', 'cg07584174', 'cg07584200', 'cg07584208', 'cg07584230', 'cg07584237', 'cg07584320', '\n",
      "1000 input query terms found no hit:\t['cg07612416', 'cg07612454', 'cg07612468', 'cg07612475', 'cg07612496', 'cg07612497', 'cg07612615', '\n",
      "1000 input query terms found no hit:\t['cg07642741', 'cg07642754', 'cg07642767', 'cg07642801', 'cg07642816', 'cg07642822', 'cg07642888', '\n",
      "1000 input query terms found no hit:\t['cg07674727', 'cg07674748', 'cg07674753', 'cg07674804', 'cg07674834', 'cg07674895', 'cg07674897', '\n",
      "1000 input query terms found no hit:\t['cg07704981', 'cg07704990', 'cg07705004', 'cg07705021', 'cg07705026', 'cg07705045', 'cg07705062', '\n",
      "1000 input query terms found no hit:\t['cg07735851', 'cg07735885', 'cg07735920', 'cg07735952', 'cg07735969', 'cg07735971', 'cg07735998', '\n",
      "1000 input query terms found no hit:\t['cg07767080', 'cg07767160', 'cg07767255', 'cg07767277', 'cg07767350', 'cg07767421', 'cg07767452', '\n",
      "1000 input query terms found no hit:\t['cg07798250', 'cg07798295', 'cg07798380', 'cg07798386', 'cg07798407', 'cg07798560', 'cg07798580', '\n",
      "1000 input query terms found no hit:\t['cg07827151', 'cg07827181', 'cg07827280', 'cg07827281', 'cg07827304', 'cg07827337', 'cg07827347', '\n",
      "1000 input query terms found no hit:\t['cg07858069', 'cg07858109', 'cg07858145', 'cg07858195', 'cg07858205', 'cg07858300', 'cg07858412', '\n",
      "1000 input query terms found no hit:\t['cg07890065', 'cg07890092', 'cg07890104', 'cg07890107', 'cg07890131', 'cg07890210', 'cg07890233', '\n",
      "1000 input query terms found no hit:\t['cg07917978', 'cg07917989', 'cg07918052', 'cg07918080', 'cg07918168', 'cg07918170', 'cg07918171', '\n",
      "1000 input query terms found no hit:\t['cg07948482', 'cg07948496', 'cg07948504', 'cg07948534', 'cg07948574', 'cg07948599', 'cg07948691', '\n",
      "1000 input query terms found no hit:\t['cg07978304', 'cg07978331', 'cg07978371', 'cg07978393', 'cg07978440', 'cg07978444', 'cg07978456', '\n",
      "1000 input query terms found no hit:\t['cg08010351', 'cg08010362', 'cg08010456', 'cg08010479', 'cg08010492', 'cg08010519', 'cg08010530', '\n",
      "1000 input query terms found no hit:\t['cg08041536', 'cg08041573', 'cg08041575', 'cg08041603', 'cg08041736', 'cg08041745', 'cg08041759', '\n",
      "1000 input query terms found no hit:\t['cg08072952', 'cg08072992', 'cg08073003', 'cg08073009', 'cg08073013', 'cg08073033', 'cg08073054', '\n",
      "1000 input query terms found no hit:\t['cg08102898', 'cg08103072', 'cg08103074', 'cg08103098', 'cg08103107', 'cg08103138', 'cg08103144', '\n",
      "1000 input query terms found no hit:\t['cg08130546', 'cg08130572', 'cg08130612', 'cg08130630', 'cg08130643', 'cg08130658', 'cg08130663', '\n",
      "1000 input query terms found no hit:\t['cg08161480', 'cg08161491', 'cg08161546', 'cg08161617', 'cg08161628', 'cg08161666', 'cg08161685', '\n",
      "1000 input query terms found no hit:\t['cg08192683', 'cg08192767', 'cg08192793', 'cg08192798', 'cg08192812', 'cg08192831', 'cg08192869', '\n",
      "1000 input query terms found no hit:\t['cg08222382', 'cg08222399', 'cg08222440', 'cg08222513', 'cg08222555', 'cg08222607', 'cg08222609', '\n",
      "1000 input query terms found no hit:\t['cg08251029', 'cg08251036', 'cg08251037', 'cg08251064', 'cg08251088', 'cg08251131', 'cg08251160', '\n",
      "1000 input query terms found no hit:\t['cg08279328', 'cg08279357', 'cg08279362', 'cg08279373', 'cg08279377', 'cg08279423', 'cg08279428', '\n",
      "1000 input query terms found no hit:\t['cg08310252', 'cg08310268', 'cg08310277', 'cg08310282', 'cg08310325', 'cg08310360', 'cg08310363', '\n",
      "1000 input query terms found no hit:\t['cg08341924', 'cg08341987', 'cg08342005', 'cg08342023', 'cg08342038', 'cg08342050', 'cg08342092', '\n",
      "1000 input query terms found no hit:\t['cg08373323', 'cg08373403', 'cg08373515', 'cg08373528', 'cg08373547', 'cg08373573', 'cg08373605', '\n",
      "1000 input query terms found no hit:\t['cg08405365', 'cg08405377', 'cg08405378', 'cg08405405', 'cg08405447', 'cg08405463', 'cg08405558', '\n",
      "1000 input query terms found no hit:\t['cg08438529', 'cg08438572', 'cg08438589', 'cg08438606', 'cg08438609', 'cg08438647', 'cg08438650', '\n",
      "1000 input query terms found no hit:\t['cg08467067', 'cg08467089', 'cg08467097', 'cg08467103', 'cg08467108', 'cg08467128', 'cg08467185', '\n",
      "1000 input query terms found no hit:\t['cg08496404', 'cg08496428', 'cg08496452', 'cg08496483', 'cg08496503', 'cg08496521', 'cg08496523', '\n",
      "1000 input query terms found no hit:\t['cg08527913', 'cg08527951', 'cg08527958', 'cg08528004', 'cg08528035', 'cg08528096', 'cg08528157', '\n",
      "1000 input query terms found no hit:\t['cg08556445', 'cg08556470', 'cg08556477', 'cg08556479', 'cg08556498', 'cg08556511', 'cg08556525', '\n",
      "1000 input query terms found no hit:\t['cg08587603', 'cg08587654', 'cg08587685', 'cg08587717', 'cg08587721', 'cg08587775', 'cg08587797', '\n",
      "1000 input query terms found no hit:\t['cg08620313', 'cg08620325', 'cg08620329', 'cg08620411', 'cg08620426', 'cg08620470', 'cg08620474', '\n",
      "1000 input query terms found no hit:\t['cg08650402', 'cg08650507', 'cg08650580', 'cg08650587', 'cg08650596', 'cg08650597', 'cg08650639', '\n",
      "1000 input query terms found no hit:\t['cg08681855', 'cg08681856', 'cg08681864', 'cg08681869', 'cg08681877', 'cg08681890', 'cg08681904', '\n",
      "1000 input query terms found no hit:\t['cg08710477', 'cg08710520', 'cg08710521', 'cg08710524', 'cg08710564', 'cg08710597', 'cg08710628', '\n",
      "1000 input query terms found no hit:\t['cg08741076', 'cg08741156', 'cg08741214', 'cg08741215', 'cg08741246', 'cg08741280', 'cg08741300', '\n",
      "1000 input query terms found no hit:\t['cg08772602', 'cg08772636', 'cg08772684', 'cg08772697', 'cg08772714', 'cg08772754', 'cg08772789', '\n",
      "1000 input query terms found no hit:\t['cg08804537', 'cg08804555', 'cg08804580', 'cg08804597', 'cg08804626', 'cg08804636', 'cg08804725', '\n",
      "1000 input query terms found no hit:\t['cg08833631', 'cg08833664', 'cg08833670', 'cg08833675', 'cg08833741', 'cg08833755', 'cg08833772', '\n",
      "1000 input query terms found no hit:\t['cg08862203', 'cg08862209', 'cg08862210', 'cg08862225', 'cg08862245', 'cg08862248', 'cg08862267', '\n",
      "1000 input query terms found no hit:\t['cg08889347', 'cg08889373', 'cg08889380', 'cg08889394', 'cg08889440', 'cg08889480', 'cg08889511', '\n",
      "1000 input query terms found no hit:\t['cg08919372', 'cg08919443', 'cg08919486', 'cg08919494', 'cg08919528', 'cg08919587', 'cg08919594', '\n",
      "1000 input query terms found no hit:\t['cg08949799', 'cg08949815', 'cg08949854', 'cg08949856', 'cg08949862', 'cg08949924', 'cg08949932', '\n",
      "1000 input query terms found no hit:\t['cg08980715', 'cg08980837', 'cg08980859', 'cg08980965', 'cg08980986', 'cg08980987', 'cg08980995', '\n",
      "1000 input query terms found no hit:\t['cg09013655', 'cg09013692', 'cg09013696', 'cg09013709', 'cg09013732', 'cg09013779', 'cg09013789', '\n",
      "1000 input query terms found no hit:\t['cg09044919', 'cg09044928', 'cg09044981', 'cg09044995', 'cg09045007', 'cg09045008', 'cg09045026', '\n",
      "1000 input query terms found no hit:\t['cg09077675', 'cg09077752', 'cg09077757', 'cg09077801', 'cg09077819', 'cg09077862', 'cg09077913', '\n",
      "1000 input query terms found no hit:\t['cg09109383', 'cg09109411', 'cg09109450', 'cg09109477', 'cg09109496', 'cg09109497', 'cg09109520', '\n",
      "1000 input query terms found no hit:\t['cg09139783', 'cg09139790', 'cg09139806', 'cg09139810', 'cg09139862', 'cg09139874', 'cg09139923', '\n",
      "1000 input query terms found no hit:\t['cg09169152', 'cg09169156', 'cg09169215', 'cg09169255', 'cg09169269', 'cg09169283', 'cg09169291', '\n",
      "1000 input query terms found no hit:\t['cg09200186', 'cg09200221', 'cg09200260', 'cg09200263', 'cg09200294', 'cg09200304', 'cg09200308', '\n",
      "1000 input query terms found no hit:\t['cg09230870', 'cg09230903', 'cg09230905', 'cg09230938', 'cg09230950', 'cg09230996', 'cg09231031', '\n",
      "1000 input query terms found no hit:\t['cg09259409', 'cg09259420', 'cg09259456', 'cg09259466', 'cg09259517', 'cg09259525', 'cg09259534', '\n",
      "1000 input query terms found no hit:\t['cg09290175', 'cg09290177', 'cg09290216', 'cg09290231', 'cg09290242', 'cg09290245', 'cg09290255', '\n",
      "1000 input query terms found no hit:\t['cg09319083', 'cg09319108', 'cg09319125', 'cg09319191', 'cg09319194', 'cg09319202', 'cg09319210', '\n",
      "1000 input query terms found no hit:\t['cg09349933', 'cg09349940', 'cg09350008', 'cg09350013', 'cg09350131', 'cg09350274', 'cg09350341', '\n",
      "1000 input query terms found no hit:\t['cg09380067', 'cg09380069', 'cg09380118', 'cg09380122', 'cg09380135', 'cg09380149', 'cg09380198', '\n",
      "1000 input query terms found no hit:\t['cg09412369', 'cg09412390', 'cg09412442', 'cg09412461', 'cg09412490', 'cg09412503', 'cg09412528', '\n",
      "1000 input query terms found no hit:\t['cg09442355', 'cg09442403', 'cg09442452', 'cg09442458', 'cg09442489', 'cg09442556', 'cg09442591', '\n",
      "1000 input query terms found no hit:\t['cg09471804', 'cg09471817', 'cg09471883', 'cg09471887', 'cg09471899', 'cg09471916', 'cg09472004', '\n",
      "1000 input query terms found no hit:\t['cg09504092', 'cg09504104', 'cg09504130', 'cg09504131', 'cg09504231', 'cg09504275', 'cg09504313', '\n",
      "1000 input query terms found no hit:\t['cg09534489', 'cg09534501', 'cg09534675', 'cg09534686', 'cg09534797', 'cg09534800', 'cg09534872', '\n",
      "1000 input query terms found no hit:\t['cg09565237', 'cg09565290', 'cg09565310', 'cg09565319', 'cg09565326', 'cg09565397', 'cg09565402', '\n",
      "1000 input query terms found no hit:\t['cg09596374', 'cg09596391', 'cg09596405', 'cg09596429', 'cg09596511', 'cg09596545', 'cg09596614', '\n",
      "1000 input query terms found no hit:\t['cg09627276', 'cg09627302', 'cg09627339', 'cg09627423', 'cg09627506', 'cg09627520', 'cg09627567', '\n",
      "1000 input query terms found no hit:\t['cg09657982', 'cg09658052', 'cg09658066', 'cg09658097', 'cg09658126', 'cg09658158', 'cg09658183', '\n",
      "1000 input query terms found no hit:\t['cg09689296', 'cg09689317', 'cg09689338', 'cg09689342', 'cg09689385', 'cg09689387', 'cg09689404', '\n",
      "1000 input query terms found no hit:\t['cg09722979', 'cg09722985', 'cg09723088', 'cg09723136', 'cg09723155', 'cg09723161', 'cg09723184', '\n",
      "1000 input query terms found no hit:\t['cg09754636', 'cg09754643', 'cg09754663', 'cg09754671', 'cg09754810', 'cg09754828', 'cg09754842', '\n",
      "1000 input query terms found no hit:\t['cg09786321', 'cg09786329', 'cg09786333', 'cg09786335', 'cg09786362', 'cg09786383', 'cg09786420', '\n",
      "1000 input query terms found no hit:\t['cg09817848', 'cg09817950', 'cg09817951', 'cg09817985', 'cg09817993', 'cg09818023', 'cg09818040', '\n",
      "1000 input query terms found no hit:\t['cg09848096', 'cg09848122', 'cg09848191', 'cg09848198', 'cg09848218', 'cg09848220', 'cg09848259', '\n",
      "1000 input query terms found no hit:\t['cg09876464', 'cg09876519', 'cg09876568', 'cg09876574', 'cg09876575', 'cg09876590', 'cg09876603', '\n",
      "1000 input query terms found no hit:\t['cg09907968', 'cg09908001', 'cg09908041', 'cg09908042', 'cg09908050', 'cg09908110', 'cg09908150', '\n",
      "1000 input query terms found no hit:\t['cg09939227', 'cg09939229', 'cg09939247', 'cg09939269', 'cg09939303', 'cg09939341', 'cg09939344', '\n",
      "1000 input query terms found no hit:\t['cg09971549', 'cg09971562', 'cg09971583', 'cg09971589', 'cg09971591', 'cg09971594', 'cg09971626', '\n",
      "1000 input query terms found no hit:\t['cg10002850', 'cg10002860', 'cg10002888', 'cg10002915', 'cg10002945', 'cg10002977', 'cg10002978', '\n",
      "1000 input query terms found no hit:\t['cg10033625', 'cg10033663', 'cg10033675', 'cg10033694', 'cg10033725', 'cg10033746', 'cg10033761', '\n",
      "1000 input query terms found no hit:\t['cg10063316', 'cg10063351', 'cg10063403', 'cg10063407', 'cg10063433', 'cg10063476', 'cg10063482', '\n",
      "1000 input query terms found no hit:\t['cg10093679', 'cg10093738', 'cg10093739', 'cg10093750', 'cg10093761', 'cg10093768', 'cg10093769', '\n",
      "1000 input query terms found no hit:\t['cg10122877', 'cg10122878', 'cg10122885', 'cg10122899', 'cg10122910', 'cg10122916', 'cg10122917', '\n",
      "1000 input query terms found no hit:\t['cg10152449', 'cg10152475', 'cg10152497', 'cg10152523', 'cg10152529', 'cg10152624', 'cg10152627', '\n",
      "1000 input query terms found no hit:\t['cg10182102', 'cg10182135', 'cg10182151', 'cg10182243', 'cg10182244', 'cg10182260', 'cg10182287', '\n",
      "1000 input query terms found no hit:\t['cg10210465', 'cg10210496', 'cg10210510', 'cg10210571', 'cg10210594', 'cg10210605', 'cg10210624', '\n",
      "1000 input query terms found no hit:\t['cg10241261', 'cg10241280', 'cg10241288', 'cg10241319', 'cg10241341', 'cg10241347', 'cg10241348', '\n",
      "1000 input query terms found no hit:\t['cg10272977', 'cg10273072', 'cg10273096', 'cg10273135', 'cg10273151', 'cg10273152', 'cg10273210', '\n",
      "1000 input query terms found no hit:\t['cg10304824', 'cg10304865', 'cg10304922', 'cg10304941', 'cg10304969', 'cg10305000', 'cg10305094', '\n",
      "1000 input query terms found no hit:\t['cg10333908', 'cg10333959', 'cg10333970', 'cg10333990', 'cg10333993', 'cg10334025', 'cg10334030', '\n",
      "1000 input query terms found no hit:\t['cg10366876', 'cg10366878', 'cg10366976', 'cg10367011', 'cg10367021', 'cg10367023', 'cg10367066', '\n",
      "1000 input query terms found no hit:\t['cg10398581', 'cg10398589', 'cg10398590', 'cg10398605', 'cg10398682', 'cg10398690', 'cg10398730', '\n",
      "1000 input query terms found no hit:\t['cg10430932', 'cg10430941', 'cg10430963', 'cg10430994', 'cg10431073', 'cg10431175', 'cg10431216', '\n",
      "1000 input query terms found no hit:\t['cg10460168', 'cg10460183', 'cg10460284', 'cg10460350', 'cg10460377', 'cg10460417', 'cg10460454', '\n",
      "1000 input query terms found no hit:\t['cg10492240', 'cg10492396', 'cg10492417', 'cg10492453', 'cg10492504', 'cg10492511', 'cg10492559', '\n",
      "1000 input query terms found no hit:\t['cg10519313', 'cg10519368', 'cg10519389', 'cg10519393', 'cg10519404', 'cg10519406', 'cg10519417', '\n",
      "1000 input query terms found no hit:\t['cg10549338', 'cg10549348', 'cg10549357', 'cg10549382', 'cg10549438', 'cg10549467', 'cg10549480', '\n",
      "1000 input query terms found no hit:\t['cg10582343', 'cg10582357', 'cg10582371', 'cg10582407', 'cg10582440', 'cg10582502', 'cg10582556', '\n",
      "1000 input query terms found no hit:\t['cg10612046', 'cg10612060', 'cg10612065', 'cg10612096', 'cg10612128', 'cg10612201', 'cg10612237', '\n",
      "1000 input query terms found no hit:\t['cg10643578', 'cg10643581', 'cg10643588', 'cg10643608', 'cg10643750', 'cg10643773', 'cg10643841', '\n",
      "1000 input query terms found no hit:\t['cg10675725', 'cg10675780', 'cg10675797', 'cg10675865', 'cg10675903', 'cg10675915', 'cg10675988', '\n",
      "1000 input query terms found no hit:\t['cg10709021', 'cg10709026', 'cg10709027', 'cg10709031', 'cg10709062', 'cg10709080', 'cg10709124', '\n",
      "1000 input query terms found no hit:\t['cg10740054', 'cg10740203', 'cg10740215', 'cg10740245', 'cg10740282', 'cg10740320', 'cg10740324', '\n",
      "1000 input query terms found no hit:\t['cg10774018', 'cg10774022', 'cg10774152', 'cg10774155', 'cg10774164', 'cg10774181', 'cg10774185', '\n",
      "1000 input query terms found no hit:\t['cg10805220', 'cg10805229', 'cg10805254', 'cg10805265', 'cg10805267', 'cg10805352', 'cg10805440', '\n",
      "1000 input query terms found no hit:\t['cg10835651', 'cg10835659', 'cg10835686', 'cg10835711', 'cg10835712', 'cg10835773', 'cg10835795', '\n",
      "1000 input query terms found no hit:\t['cg10868218', 'cg10868235', 'cg10868277', 'cg10868293', 'cg10868297', 'cg10868351', 'cg10868379', '\n",
      "1000 input query terms found no hit:\t['cg10901031', 'cg10901059', 'cg10901099', 'cg10901134', 'cg10901187', 'cg10901280', 'cg10901291', '\n",
      "1000 input query terms found no hit:\t['cg10933494', 'cg10933497', 'cg10933529', 'cg10933573', 'cg10933616', 'cg10933625', 'cg10933684', '\n",
      "1000 input query terms found no hit:\t['cg10965117', 'cg10965131', 'cg10965145', 'cg10965164', 'cg10965178', 'cg10965183', 'cg10965259', '\n",
      "1000 input query terms found no hit:\t['cg10996039', 'cg10996058', 'cg10996109', 'cg10996111', 'cg10996127', 'cg10996143', 'cg10996148', '\n",
      "1000 input query terms found no hit:\t['cg11028730', 'cg11028738', 'cg11028747', 'cg11028751', 'cg11028755', 'cg11028770', 'cg11028872', '\n",
      "1000 input query terms found no hit:\t['cg11062167', 'cg11062168', 'cg11062287', 'cg11062387', 'cg11062405', 'cg11062417', 'cg11062418', '\n",
      "1000 input query terms found no hit:\t['cg11094568', 'cg11094656', 'cg11094660', 'cg11094694', 'cg11094705', 'cg11094740', 'cg11094839', '\n",
      "1000 input query terms found no hit:\t['cg11127021', 'cg11127022', 'cg11127112', 'cg11127124', 'cg11127156', 'cg11127164', 'cg11127249', '\n",
      "1000 input query terms found no hit:\t['cg11158587', 'cg11158613', 'cg11158629', 'cg11158641', 'cg11158729', 'cg11158760', 'cg11158819', '\n",
      "1000 input query terms found no hit:\t['cg11192053', 'cg11192055', 'cg11192059', 'cg11192090', 'cg11192140', 'cg11192141', 'cg11192169', '\n",
      "1000 input query terms found no hit:\t['cg11221537', 'cg11221598', 'cg11221635', 'cg11221657', 'cg11221687', 'cg11221741', 'cg11221800', '\n",
      "1000 input query terms found no hit:\t['cg11251293', 'cg11251309', 'cg11251319', 'cg11251367', 'cg11251378', 'cg11251399', 'cg11251470', '\n",
      "1000 input query terms found no hit:\t['cg11284147', 'cg11284196', 'cg11284197', 'cg11284281', 'cg11284299', 'cg11284316', 'cg11284329', '\n",
      "1000 input query terms found no hit:\t['cg11314687', 'cg11314748', 'cg11314757', 'cg11314779', 'cg11314787', 'cg11314818', 'cg11314819', '\n",
      "1000 input query terms found no hit:\t['cg11345718', 'cg11345719', 'cg11345811', 'cg11345899', 'cg11345909', 'cg11345918', 'cg11345955', '\n",
      "1000 input query terms found no hit:\t['cg11377279', 'cg11377286', 'cg11377293', 'cg11377298', 'cg11377336', 'cg11377397', 'cg11377440', '\n",
      "1000 input query terms found no hit:\t['cg11408657', 'cg11408710', 'cg11408729', 'cg11408754', 'cg11408771', 'cg11408812', 'cg11408879', '\n",
      "1000 input query terms found no hit:\t['cg11440240', 'cg11440317', 'cg11440344', 'cg11440375', 'cg11440409', 'cg11440424', 'cg11440444', '\n",
      "1000 input query terms found no hit:\t['cg11473757', 'cg11473771', 'cg11473803', 'cg11473855', 'cg11473856', 'cg11473876', 'cg11474075', '\n",
      "1000 input query terms found no hit:\t['cg11508532', 'cg11508568', 'cg11508571', 'cg11508593', 'cg11508595', 'cg11508674', 'cg11508714', '\n",
      "1000 input query terms found no hit:\t['cg11541587', 'cg11541601', 'cg11541630', 'cg11541668', 'cg11541678', 'cg11541712', 'cg11541758', '\n",
      "1000 input query terms found no hit:\t['cg11574580', 'cg11574612', 'cg11574620', 'cg11574645', 'cg11574665', 'cg11574745', 'cg11574758', '\n",
      "1000 input query terms found no hit:\t['cg11606215', 'cg11606229', 'cg11606261', 'cg11606268', 'cg11606286', 'cg11606295', 'cg11606365', '\n",
      "1000 input query terms found no hit:\t['cg11638921', 'cg11638962', 'cg11638966', 'cg11638980', 'cg11638982', 'cg11638998', 'cg11639001', '\n",
      "1000 input query terms found no hit:\t['cg11668794', 'cg11668797', 'cg11668800', 'cg11668813', 'cg11668822', 'cg11668844', 'cg11668876', '\n",
      "1000 input query terms found no hit:\t['cg11700735', 'cg11700800', 'cg11700806', 'cg11700824', 'cg11700868', 'cg11700878', 'cg11700897', '\n",
      "1000 input query terms found no hit:\t['cg11730722', 'cg11730826', 'cg11730925', 'cg11730930', 'cg11730958', 'cg11730961', 'cg11731013', '\n",
      "1000 input query terms found no hit:\t['cg11760356', 'cg11760361', 'cg11760363', 'cg11760389', 'cg11760395', 'cg11760414', 'cg11760500', '\n",
      "1000 input query terms found no hit:\t['cg11792664', 'cg11792671', 'cg11792728', 'cg11792766', 'cg11792771', 'cg11792779', 'cg11792786', '\n",
      "1000 input query terms found no hit:\t['cg11821738', 'cg11821759', 'cg11821795', 'cg11821796', 'cg11821817', 'cg11821824', 'cg11821865', '\n",
      "1000 input query terms found no hit:\t['cg11851653', 'cg11851656', 'cg11851753', 'cg11851772', 'cg11851790', 'cg11851792', 'cg11851804', '\n",
      "1000 input query terms found no hit:\t['cg11881934', 'cg11881963', 'cg11881989', 'cg11881990', 'cg11882042', 'cg11882053', 'cg11882101', '\n",
      "1000 input query terms found no hit:\t['cg11913696', 'cg11913753', 'cg11913867', 'cg11913894', 'cg11913917', 'cg11913951', 'cg11914031', '\n",
      "1000 input query terms found no hit:\t['cg11945507', 'cg11945538', 'cg11945554', 'cg11945565', 'cg11945604', 'cg11945644', 'cg11945668', '\n",
      "1000 input query terms found no hit:\t['cg11977443', 'cg11977488', 'cg11977491', 'cg11977500', 'cg11977562', 'cg11977605', 'cg11977618', '\n",
      "1000 input query terms found no hit:\t['cg12010373', 'cg12010396', 'cg12010413', 'cg12010452', 'cg12010464', 'cg12010476', 'cg12010544', '\n",
      "1000 input query terms found no hit:\t['cg12042598', 'cg12042629', 'cg12042633', 'cg12042678', 'cg12042689', 'cg12042711', 'cg12042714', '\n",
      "1000 input query terms found no hit:\t['cg12072740', 'cg12072757', 'cg12072770', 'cg12072789', 'cg12072803', 'cg12072822', 'cg12072830', '\n",
      "1000 input query terms found no hit:\t['cg12102702', 'cg12102756', 'cg12102838', 'cg12102973', 'cg12103037', 'cg12103052', 'cg12103149', '\n",
      "1000 input query terms found no hit:\t['cg12134754', 'cg12134755', 'cg12134798', 'cg12134806', 'cg12134886', 'cg12134894', 'cg12134900', '\n",
      "1000 input query terms found no hit:\t['cg12166313', 'cg12166348', 'cg12166435', 'cg12166447', 'cg12166476', 'cg12166502', 'cg12166519', '\n",
      "1000 input query terms found no hit:\t['cg12197380', 'cg12197399', 'cg12197421', 'cg12197457', 'cg12197459', 'cg12197470', 'cg12197557', '\n",
      "1000 input query terms found no hit:\t['cg12229644', 'cg12229666', 'cg12229727', 'cg12229764', 'cg12229767', 'cg12229775', 'cg12229832', '\n",
      "1000 input query terms found no hit:\t['cg12262610', 'cg12262617', 'cg12262650', 'cg12262681', 'cg12262690', 'cg12262698', 'cg12262749', '\n",
      "1000 input query terms found no hit:\t['cg12297570', 'cg12297590', 'cg12297619', 'cg12297625', 'cg12297652', 'cg12297676', 'cg12297684', '\n",
      "1000 input query terms found no hit:\t['cg12331896', 'cg12331918', 'cg12331932', 'cg12331980', 'cg12331986', 'cg12332073', 'cg12332083', '\n",
      "1000 input query terms found no hit:\t['cg12366345', 'cg12366423', 'cg12366450', 'cg12366453', 'cg12366582', 'cg12366597', 'cg12366623', '\n",
      "1000 input query terms found no hit:\t['cg12400041', 'cg12400070', 'cg12400097', 'cg12400102', 'cg12400114', 'cg12400119', 'cg12400121', '\n",
      "1000 input query terms found no hit:\t['cg12431401', 'cg12431453', 'cg12431483', 'cg12431496', 'cg12431498', 'cg12431505', 'cg12431517', '\n",
      "1000 input query terms found no hit:\t['cg12460233', 'cg12460254', 'cg12460281', 'cg12460307', 'cg12460322', 'cg12460362', 'cg12460375', '\n",
      "1000 input query terms found no hit:\t['cg12491805', 'cg12491821', 'cg12491837', 'cg12491839', 'cg12491853', 'cg12491966', 'cg12491991', '\n",
      "1000 input query terms found no hit:\t['cg12523924', 'cg12523932', 'cg12523974', 'cg12524061', 'cg12524068', 'cg12524168', 'cg12524179', '\n",
      "1000 input query terms found no hit:\t['cg12556601', 'cg12556616', 'cg12556635', 'cg12556644', 'cg12556775', 'cg12556802', 'cg12556823', '\n",
      "1000 input query terms found no hit:\t['cg12588678', 'cg12588699', 'cg12588704', 'cg12588725', 'cg12588743', 'cg12588763', 'cg12588834', '\n",
      "1000 input query terms found no hit:\t['cg12618270', 'cg12618475', 'cg12618499', 'cg12618546', 'cg12618571', 'cg12618574', 'cg12618624', '\n",
      "1000 input query terms found no hit:\t['cg12650011', 'cg12650153', 'cg12650227', 'cg12650252', 'cg12650265', 'cg12650267', 'cg12650305', '\n",
      "1000 input query terms found no hit:\t['cg12682805', 'cg12682817', 'cg12682862', 'cg12682872', 'cg12682914', 'cg12682916', 'cg12682931', '\n",
      "1000 input query terms found no hit:\t['cg12716056', 'cg12716083', 'cg12716200', 'cg12716207', 'cg12716232', 'cg12716233', 'cg12716235', '\n",
      "1000 input query terms found no hit:\t['cg12749360', 'cg12749396', 'cg12749428', 'cg12749439', 'cg12749467', 'cg12749468', 'cg12749531', '\n",
      "1000 input query terms found no hit:\t['cg12783942', 'cg12783943', 'cg12783951', 'cg12784049', 'cg12784089', 'cg12784104', 'cg12784120', '\n",
      "1000 input query terms found no hit:\t['cg12815355', 'cg12815359', 'cg12815372', 'cg12815381', 'cg12815555', 'cg12815605', 'cg12815656', '\n",
      "1000 input query terms found no hit:\t['cg12848571', 'cg12848592', 'cg12848614', 'cg12848627', 'cg12848631', 'cg12848639', 'cg12848697', '\n",
      "1000 input query terms found no hit:\t['cg12881363', 'cg12881368', 'cg12881393', 'cg12881443', 'cg12881453', 'cg12881469', 'cg12881470', '\n",
      "1000 input query terms found no hit:\t['cg12913344', 'cg12913429', 'cg12913434', 'cg12913479', 'cg12913494', 'cg12913496', 'cg12913502', '\n",
      "1000 input query terms found no hit:\t['cg12944870', 'cg12944923', 'cg12944944', 'cg12944966', 'cg12945002', 'cg12945054', 'cg12945092', '\n",
      "1000 input query terms found no hit:\t['cg12978308', 'cg12978357', 'cg12978433', 'cg12978539', 'cg12978573', 'cg12978575', 'cg12978581', '\n",
      "1000 input query terms found no hit:\t['cg13012115', 'cg13012163', 'cg13012206', 'cg13012245', 'cg13012248', 'cg13012325', 'cg13012330', '\n",
      "1000 input query terms found no hit:\t['cg13045501', 'cg13045504', 'cg13045514', 'cg13045540', 'cg13045555', 'cg13045593', 'cg13045611', '\n",
      "1000 input query terms found no hit:\t['cg13077138', 'cg13077150', 'cg13077160', 'cg13077165', 'cg13077174', 'cg13077175', 'cg13077190', '\n",
      "1000 input query terms found no hit:\t['cg13108609', 'cg13108646', 'cg13108653', 'cg13108681', 'cg13108687', 'cg13108852', 'cg13108869', '\n",
      "1000 input query terms found no hit:\t['cg13142499', 'cg13142612', 'cg13142689', 'cg13142700', 'cg13142714', 'cg13142750', 'cg13142917', '\n",
      "1000 input query terms found no hit:\t['cg13174077', 'cg13174083', 'cg13174084', 'cg13174161', 'cg13174197', 'cg13174229', 'cg13174253', '\n",
      "1000 input query terms found no hit:\t['cg13207333', 'cg13207339', 'cg13207340', 'cg13207345', 'cg13207467', 'cg13207490', 'cg13207501', '\n",
      "1000 input query terms found no hit:\t['cg13239420', 'cg13239483', 'cg13239514', 'cg13239540', 'cg13239563', 'cg13239572', 'cg13239599', '\n",
      "1000 input query terms found no hit:\t['cg13271939', 'cg13271951', 'cg13271963', 'cg13272029', 'cg13272096', 'cg13272108', 'cg13272119', '\n",
      "1000 input query terms found no hit:\t['cg13302229', 'cg13302309', 'cg13302346', 'cg13302369', 'cg13302386', 'cg13302401', 'cg13302409', '\n",
      "1000 input query terms found no hit:\t['cg13334579', 'cg13334650', 'cg13334673', 'cg13334687', 'cg13334727', 'cg13334757', 'cg13334788', '\n",
      "1000 input query terms found no hit:\t['cg13365901', 'cg13365972', 'cg13366048', 'cg13366057', 'cg13366111', 'cg13366112', 'cg13366181', '\n",
      "1000 input query terms found no hit:\t['cg13395215', 'cg13395333', 'cg13395344', 'cg13395373', 'cg13395410', 'cg13395516', 'cg13395568', '\n",
      "1000 input query terms found no hit:\t['cg13425391', 'cg13425422', 'cg13425478', 'cg13425489', 'cg13425498', 'cg13425515', 'cg13425605', '\n",
      "1000 input query terms found no hit:\t['cg13453175', 'cg13453193', 'cg13453196', 'cg13453203', 'cg13453228', 'cg13453244', 'cg13453253', '\n",
      "1000 input query terms found no hit:\t['cg13481379', 'cg13481413', 'cg13481455', 'cg13481522', 'cg13481590', 'cg13481638', 'cg13481734', '\n",
      "1000 input query terms found no hit:\t['cg13510050', 'cg13510054', 'cg13510110', 'cg13510114', 'cg13510176', 'cg13510262', 'cg13510318', '\n",
      "1000 input query terms found no hit:\t['cg13537578', 'cg13537590', 'cg13537595', 'cg13537646', 'cg13537662', 'cg13537679', 'cg13537752', '\n",
      "1000 input query terms found no hit:\t['cg13565382', 'cg13565400', 'cg13565401', 'cg13565461', 'cg13565468', 'cg13565492', 'cg13565499', '\n",
      "1000 input query terms found no hit:\t['cg13593284', 'cg13593287', 'cg13593297', 'cg13593333', 'cg13593363', 'cg13593376', 'cg13593391', '\n",
      "1000 input query terms found no hit:\t['cg13620744', 'cg13620770', 'cg13620799', 'cg13620808', 'cg13620814', 'cg13620881', 'cg13620921', '\n",
      "1000 input query terms found no hit:\t['cg13649975', 'cg13650024', 'cg13650062', 'cg13650099', 'cg13650104', 'cg13650137', 'cg13650143', '\n",
      "1000 input query terms found no hit:\t['cg13679303', 'cg13679311', 'cg13679443', 'cg13679466', 'cg13679478', 'cg13679619', 'cg13679679', '\n",
      "1000 input query terms found no hit:\t['cg13707123', 'cg13707183', 'cg13707189', 'cg13707194', 'cg13707224', 'cg13707335', 'cg13707337', '\n",
      "1000 input query terms found no hit:\t['cg13734941', 'cg13734948', 'cg13734953', 'cg13734972', 'cg13734996', 'cg13735018', 'cg13735166', '\n",
      "1000 input query terms found no hit:\t['cg13765961', 'cg13765978', 'cg13766027', 'cg13766030', 'cg13766031', 'cg13766043', 'cg13766044', '\n",
      "1000 input query terms found no hit:\t['cg13795321', 'cg13795336', 'cg13795379', 'cg13795448', 'cg13795465', 'cg13795501', 'cg13795558', '\n",
      "1000 input query terms found no hit:\t['cg13823621', 'cg13823627', 'cg13823643', 'cg13823666', 'cg13823701', 'cg13823708', 'cg13823791', '\n",
      "1000 input query terms found no hit:\t['cg13854030', 'cg13854035', 'cg13854043', 'cg13854081', 'cg13854092', 'cg13854135', 'cg13854193', '\n",
      "1000 input query terms found no hit:\t['cg13883113', 'cg13883127', 'cg13883173', 'cg13883186', 'cg13883202', 'cg13883232', 'cg13883256', '\n",
      "1000 input query terms found no hit:\t['cg13911772', 'cg13911785', 'cg13911801', 'cg13911828', 'cg13911847', 'cg13911857', 'cg13911859', '\n",
      "1000 input query terms found no hit:\t['cg13938257', 'cg13938341', 'cg13938349', 'cg13938361', 'cg13938376', 'cg13938410', 'cg13938467', '\n",
      "1000 input query terms found no hit:\t['cg13967723', 'cg13967780', 'cg13967811', 'cg13967864', 'cg13967881', 'cg13967908', 'cg13967936', '\n",
      "1000 input query terms found no hit:\t['cg13995325', 'cg13995374', 'cg13995389', 'cg13995399', 'cg13995427', 'cg13995432', 'cg13995453', '\n",
      "1000 input query terms found no hit:\t['cg14022022', 'cg14022057', 'cg14022090', 'cg14022105', 'cg14022137', 'cg14022144', 'cg14022183', '\n",
      "1000 input query terms found no hit:\t['cg14049380', 'cg14049382', 'cg14049391', 'cg14049461', 'cg14049470', 'cg14049535', 'cg14049549', '\n",
      "1000 input query terms found no hit:\t['cg14079785', 'cg14079845', 'cg14079867', 'cg14079893', 'cg14080001', 'cg14080033', 'cg14080048', '\n",
      "1000 input query terms found no hit:\t['cg14107457', 'cg14107485', 'cg14107488', 'cg14107561', 'cg14107583', 'cg14107638', 'cg14107644', '\n",
      "1000 input query terms found no hit:\t['cg14137548', 'cg14137558', 'cg14137578', 'cg14137606', 'cg14137617', 'cg14137625', 'cg14137646', '\n",
      "1000 input query terms found no hit:\t['cg14166459', 'cg14166485', 'cg14166494', 'cg14166518', 'cg14166520', 'cg14166625', 'cg14166629', '\n",
      "1000 input query terms found no hit:\t['cg14194121', 'cg14194187', 'cg14194192', 'cg14194253', 'cg14194263', 'cg14194287', 'cg14194317', '\n",
      "1000 input query terms found no hit:\t['cg14220722', 'cg14220726', 'cg14220764', 'cg14220777', 'cg14220830', 'cg14220884', 'cg14220887', '\n",
      "1000 input query terms found no hit:\t['cg14249120', 'cg14249147', 'cg14249148', 'cg14249155', 'cg14249168', 'cg14249174', 'cg14249212', '\n",
      "1000 input query terms found no hit:\t['cg14277924', 'cg14277925', 'cg14277989', 'cg14278029', 'cg14278054', 'cg14278057', 'cg14278073', '\n",
      "1000 input query terms found no hit:\t['cg14306023', 'cg14306058', 'cg14306094', 'cg14306118', 'cg14306145', 'cg14306161', 'cg14306254', '\n",
      "1000 input query terms found no hit:\t['cg14334041', 'cg14334069', 'cg14334109', 'cg14334147', 'cg14334161', 'cg14334165', 'cg14334173', '\n",
      "1000 input query terms found no hit:\t['cg14363533', 'cg14363563', 'cg14363604', 'cg14363607', 'cg14363636', 'cg14363661', 'cg14363714', '\n",
      "1000 input query terms found no hit:\t['cg14392253', 'cg14392283', 'cg14392292', 'cg14392344', 'cg14392387', 'cg14392392', 'cg14392476', '\n",
      "1000 input query terms found no hit:\t['cg14420490', 'cg14420494', 'cg14420496', 'cg14420550', 'cg14420567', 'cg14420589', 'cg14420612', '\n",
      "1000 input query terms found no hit:\t['cg14449059', 'cg14449085', 'cg14449180', 'cg14449209', 'cg14449251', 'cg14449256', 'cg14449263', '\n",
      "1000 input query terms found no hit:\t['cg14477615', 'cg14477619', 'cg14477622', 'cg14477673', 'cg14477708', 'cg14477767', 'cg14477768', '\n",
      "1000 input query terms found no hit:\t['cg14506970', 'cg14507002', 'cg14507014', 'cg14507016', 'cg14507042', 'cg14507062', 'cg14507077', '\n",
      "1000 input query terms found no hit:\t['cg14533336', 'cg14533361', 'cg14533390', 'cg14533401', 'cg14533412', 'cg14533445', 'cg14533461', '\n",
      "1000 input query terms found no hit:\t['cg14560131', 'cg14560133', 'cg14560144', 'cg14560226', 'cg14560240', 'cg14560241', 'cg14560276', '\n",
      "1000 input query terms found no hit:\t['cg14588411', 'cg14588422', 'cg14588439', 'cg14588448', 'cg14588457', 'cg14588463', 'cg14588512', '\n",
      "1000 input query terms found no hit:\t['cg14616761', 'cg14616819', 'cg14616824', 'cg14616881', 'cg14616910', 'cg14616942', 'cg14617010', '\n",
      "1000 input query terms found no hit:\t['cg14646893', 'cg14646944', 'cg14646963', 'cg14646974', 'cg14646977', 'cg14646983', 'cg14646998', '\n",
      "1000 input query terms found no hit:\t['cg14676529', 'cg14676592', 'cg14676666', 'cg14676702', 'cg14676711', 'cg14676794', 'cg14676811', '\n",
      "1000 input query terms found no hit:\t['cg14708076', 'cg14708080', 'cg14708132', 'cg14708218', 'cg14708284', 'cg14708311', 'cg14708333', '\n",
      "1000 input query terms found no hit:\t['cg14741685', 'cg14741691', 'cg14741713', 'cg14741792', 'cg14741795', 'cg14741804', 'cg14741811', '\n",
      "1000 input query terms found no hit:\t['cg14774044', 'cg14774045', 'cg14774061', 'cg14774086', 'cg14774117', 'cg14774134', 'cg14774165', '\n",
      "1000 input query terms found no hit:\t['cg14803632', 'cg14803663', 'cg14803703', 'cg14803733', 'cg14803762', 'cg14803765', 'cg14803789', '\n",
      "1000 input query terms found no hit:\t['cg14835473', 'cg14835484', 'cg14835517', 'cg14835529', 'cg14835545', 'cg14835575', 'cg14835701', '\n",
      "1000 input query terms found no hit:\t['cg14864773', 'cg14864782', 'cg14864829', 'cg14864852', 'cg14864853', 'cg14864856', 'cg14864895', '\n",
      "1000 input query terms found no hit:\t['cg14897425', 'cg14897471', 'cg14897563', 'cg14897569', 'cg14897683', 'cg14897752', 'cg14897765', '\n",
      "1000 input query terms found no hit:\t['cg14929757', 'cg14929805', 'cg14929883', 'cg14929913', 'cg14929982', 'cg14930000', 'cg14930059', '\n",
      "1000 input query terms found no hit:\t['cg14962584', 'cg14962610', 'cg14962616', 'cg14962628', 'cg14962754', 'cg14962774', 'cg14962776', '\n",
      "1000 input query terms found no hit:\t['cg14994935', 'cg14994944', 'cg14994947', 'cg14994986', 'cg14995018', 'cg14995036', 'cg14995047', '\n",
      "1000 input query terms found no hit:\t['cg15027295', 'cg15027300', 'cg15027303', 'cg15027328', 'cg15027358', 'cg15027359', 'cg15027367', '\n",
      "1000 input query terms found no hit:\t['cg15058799', 'cg15058836', 'cg15058852', 'cg15058869', 'cg15058894', 'cg15058901', 'cg15058905', '\n",
      "1000 input query terms found no hit:\t['cg15088602', 'cg15088646', 'cg15088648', 'cg15088679', 'cg15088708', 'cg15088752', 'cg15088759', '\n",
      "1000 input query terms found no hit:\t['cg15120283', 'cg15120315', 'cg15120333', 'cg15120341', 'cg15120380', 'cg15120413', 'cg15120426', '\n",
      "1000 input query terms found no hit:\t['cg15151378', 'cg15151387', 'cg15151492', 'cg15151654', 'cg15151670', 'cg15151685', 'cg15151709', '\n",
      "1000 input query terms found no hit:\t['cg15184386', 'cg15184393', 'cg15184452', 'cg15184495', 'cg15184507', 'cg15184529', 'cg15184602', '\n",
      "1000 input query terms found no hit:\t['cg15218967', 'cg15218995', 'cg15219014', 'cg15219029', 'cg15219125', 'cg15219134', 'cg15219145', '\n",
      "1000 input query terms found no hit:\t['cg15251634', 'cg15251675', 'cg15251679', 'cg15251697', 'cg15251729', 'cg15251748', 'cg15251779', '\n",
      "1000 input query terms found no hit:\t['cg15286748', 'cg15286784', 'cg15286847', 'cg15286905', 'cg15286942', 'cg15286945', 'cg15286970', '\n",
      "1000 input query terms found no hit:\t['cg15321814', 'cg15321908', 'cg15321915', 'cg15321957', 'cg15321979', 'cg15321997', 'cg15322070', '\n",
      "1000 input query terms found no hit:\t['cg15355387', 'cg15355388', 'cg15355395', 'cg15355400', 'cg15355401', 'cg15355406', 'cg15355459', '\n",
      "1000 input query terms found no hit:\t['cg15389191', 'cg15389199', 'cg15389227', 'cg15389265', 'cg15389338', 'cg15389373', 'cg15389392', '\n",
      "1000 input query terms found no hit:\t['cg15420468', 'cg15420515', 'cg15420606', 'cg15420615', 'cg15420634', 'cg15420678', 'cg15420683', '\n",
      "1000 input query terms found no hit:\t['cg15452952', 'cg15452970', 'cg15452973', 'cg15452985', 'cg15453007', 'cg15453092', 'cg15453097', '\n",
      "1000 input query terms found no hit:\t['cg15484742', 'cg15484803', 'cg15484811', 'cg15484899', 'cg15484904', 'cg15484920', 'cg15484923', '\n",
      "1000 input query terms found no hit:\t['cg15521097', 'cg15521256', 'cg15521264', 'cg15521289', 'cg15521326', 'cg15521382', 'cg15521442', '\n",
      "1000 input query terms found no hit:\t['cg15554378', 'cg15554401', 'cg15554410', 'cg15554421', 'cg15554429', 'cg15554438', 'cg15554463', '\n",
      "1000 input query terms found no hit:\t['cg15585190', 'cg15585213', 'cg15585230', 'cg15585264', 'cg15585294', 'cg15585318', 'cg15585324', '\n",
      "1000 input query terms found no hit:\t['cg15617545', 'cg15617548', 'cg15617549', 'cg15617591', 'cg15617609', 'cg15617639', 'cg15617640', '\n",
      "1000 input query terms found no hit:\t['cg15648058', 'cg15648060', 'cg15648081', 'cg15648124', 'cg15648188', 'cg15648207', 'cg15648239', '\n",
      "1000 input query terms found no hit:\t['cg15679784', 'cg15679799', 'cg15679813', 'cg15679829', 'cg15679841', 'cg15679849', 'cg15679892', '\n",
      "1000 input query terms found no hit:\t['cg15710264', 'cg15710370', 'cg15710437', 'cg15710462', 'cg15710478', 'cg15710545', 'cg15710554', '\n",
      "1000 input query terms found no hit:\t['cg15739543', 'cg15739546', 'cg15739551', 'cg15739563', 'cg15739570', 'cg15739581', 'cg15739591', '\n",
      "1000 input query terms found no hit:\t['cg15774334', 'cg15774351', 'cg15774391', 'cg15774437', 'cg15774458', 'cg15774465', 'cg15774479', '\n",
      "1000 input query terms found no hit:\t['cg15807154', 'cg15807180', 'cg15807181', 'cg15807202', 'cg15807287', 'cg15807288', 'cg15807314', '\n",
      "1000 input query terms found no hit:\t['cg15835542', 'cg15835620', 'cg15835644', 'cg15835657', 'cg15835663', 'cg15835664', 'cg15835713', '\n",
      "1000 input query terms found no hit:\t['cg15865679', 'cg15865685', 'cg15865690', 'cg15865714', 'cg15865722', 'cg15865771', 'cg15865827', '\n",
      "1000 input query terms found no hit:\t['cg15898494', 'cg15898526', 'cg15898536', 'cg15898604', 'cg15898859', 'cg15898874', 'cg15898936', '\n",
      "1000 input query terms found no hit:\t['cg15929980', 'cg15930025', 'cg15930041', 'cg15930076', 'cg15930085', 'cg15930120', 'cg15930145', '\n",
      "1000 input query terms found no hit:\t['cg15963983', 'cg15963988', 'cg15964018', 'cg15964066', 'cg15964080', 'cg15964132', 'cg15964153', '\n",
      "1000 input query terms found no hit:\t['cg15994910', 'cg15994971', 'cg15994981', 'cg15994995', 'cg15994998', 'cg15994999', 'cg15995070', '\n",
      "1000 input query terms found no hit:\t['cg16023796', 'cg16023864', 'cg16023885', 'cg16023894', 'cg16023896', 'cg16023912', 'cg16023914', '\n",
      "1000 input query terms found no hit:\t['cg16056219', 'cg16056272', 'cg16056311', 'cg16056334', 'cg16056398', 'cg16056449', 'cg16056465', '\n",
      "1000 input query terms found no hit:\t['cg16088921', 'cg16088930', 'cg16088971', 'cg16088978', 'cg16089004', 'cg16089026', 'cg16089031', '\n",
      "1000 input query terms found no hit:\t['cg16122106', 'cg16122110', 'cg16122133', 'cg16122145', 'cg16122180', 'cg16122269', 'cg16122287', '\n",
      "1000 input query terms found no hit:\t['cg16153388', 'cg16153400', 'cg16153463', 'cg16153547', 'cg16153549', 'cg16153601', 'cg16153625', '\n",
      "1000 input query terms found no hit:\t['cg16185501', 'cg16185521', 'cg16185523', 'cg16185626', 'cg16185636', 'cg16185661', 'cg16185666', '\n",
      "1000 input query terms found no hit:\t['cg16217885', 'cg16217908', 'cg16218009', 'cg16218069', 'cg16218074', 'cg16218089', 'cg16218100', '\n",
      "1000 input query terms found no hit:\t['cg16248421', 'cg16248432', 'cg16248435', 'cg16248448', 'cg16248511', 'cg16248515', 'cg16248554', '\n",
      "1000 input query terms found no hit:\t['cg16277843', 'cg16277868', 'cg16277885', 'cg16277915', 'cg16277922', 'cg16277944', 'cg16277998', '\n",
      "1000 input query terms found no hit:\t['cg16310145', 'cg16310192', 'cg16310215', 'cg16310262', 'cg16310313', 'cg16310322', 'cg16310388', '\n",
      "1000 input query terms found no hit:\t['cg16339286', 'cg16339370', 'cg16339402', 'cg16339434', 'cg16339435', 'cg16339541', 'cg16339550', '\n",
      "1000 input query terms found no hit:\t['cg16370778', 'cg16370795', 'cg16370826', 'cg16370863', 'cg16370869', 'cg16370875', 'cg16370882', '\n",
      "1000 input query terms found no hit:\t['cg16401457', 'cg16401465', 'cg16401490', 'cg16401529', 'cg16401578', 'cg16401608', 'cg16401668', '\n",
      "1000 input query terms found no hit:\t['cg16430166', 'cg16430184', 'cg16430215', 'cg16430256', 'cg16430267', 'cg16430279', 'cg16430295', '\n",
      "1000 input query terms found no hit:\t['cg16461350', 'cg16461416', 'cg16461423', 'cg16461427', 'cg16461485', 'cg16461522', 'cg16461530', '\n",
      "1000 input query terms found no hit:\t['cg16494291', 'cg16494337', 'cg16494350', 'cg16494397', 'cg16494406', 'cg16494429', 'cg16494448', '\n",
      "1000 input query terms found no hit:\t['cg16524312', 'cg16524331', 'cg16524332', 'cg16524353', 'cg16524385', 'cg16524492', 'cg16524558', '\n",
      "1000 input query terms found no hit:\t['cg16553238', 'cg16553272', 'cg16553297', 'cg16553327', 'cg16553338', 'cg16553354', 'cg16553414', '\n",
      "1000 input query terms found no hit:\t['cg16583560', 'cg16583572', 'cg16583598', 'cg16583672', 'cg16583745', 'cg16583817', 'cg16583863', '\n",
      "1000 input query terms found no hit:\t['cg16614288', 'cg16614307', 'cg16614313', 'cg16614332', 'cg16614335', 'cg16614484', 'cg16614489', '\n",
      "1000 input query terms found no hit:\t['cg16644567', 'cg16644642', 'cg16644647', 'cg16644760', 'cg16644764', 'cg16644803', 'cg16644967', '\n",
      "1000 input query terms found no hit:\t['cg16672770', 'cg16672782', 'cg16672818', 'cg16672830', 'cg16672904', 'cg16672922', 'cg16672925', '\n",
      "1000 input query terms found no hit:\t['cg16703410', 'cg16703413', 'cg16703420', 'cg16703427', 'cg16703456', 'cg16703466', 'cg16703508', '\n",
      "1000 input query terms found no hit:\t['cg16731811', 'cg16731907', 'cg16731958', 'cg16731970', 'cg16732013', 'cg16732046', 'cg16732077', '\n",
      "1000 input query terms found no hit:\t['cg16764274', 'cg16764296', 'cg16764306', 'cg16764342', 'cg16764362', 'cg16764368', 'cg16764389', '\n",
      "1000 input query terms found no hit:\t['cg16796075', 'cg16796091', 'cg16796096', 'cg16796107', 'cg16796127', 'cg16796135', 'cg16796136', '\n",
      "1000 input query terms found no hit:\t['cg16830379', 'cg16830475', 'cg16830479', 'cg16830504', 'cg16830520', 'cg16830523', 'cg16830527', '\n",
      "1000 input query terms found no hit:\t['cg16864411', 'cg16864418', 'cg16864444', 'cg16864490', 'cg16864513', 'cg16864587', 'cg16864589', '\n",
      "1000 input query terms found no hit:\t['cg16898491', 'cg16898495', 'cg16898549', 'cg16898576', 'cg16898619', 'cg16898621', 'cg16898818', '\n",
      "1000 input query terms found no hit:\t['cg16928523', 'cg16928526', 'cg16928551', 'cg16928657', 'cg16928752', 'cg16928776', 'cg16928795', '\n",
      "1000 input query terms found no hit:\t['cg16962216', 'cg16962259', 'cg16962314', 'cg16962341', 'cg16962392', 'cg16962403', 'cg16962412', '\n",
      "1000 input query terms found no hit:\t['cg16996416', 'cg16996442', 'cg16996457', 'cg16996544', 'cg16996571', 'cg16996574', 'cg16996575', '\n",
      "1000 input query terms found no hit:\t['cg17026088', 'cg17026141', 'cg17026151', 'cg17026162', 'cg17026220', 'cg17026243', 'cg17026245', '\n",
      "1000 input query terms found no hit:\t['cg17058012', 'cg17058125', 'cg17058288', 'cg17058296', 'cg17058383', 'cg17058475', 'cg17058565', '\n",
      "1000 input query terms found no hit:\t['cg17090989', 'cg17091020', 'cg17091024', 'cg17091049', 'cg17091056', 'cg17091176', 'cg17091196', '\n",
      "1000 input query terms found no hit:\t['cg17123420', 'cg17123468', 'cg17123508', 'cg17123533', 'cg17123534', 'cg17123539', 'cg17123560', '\n",
      "1000 input query terms found no hit:\t['cg17153733', 'cg17153776', 'cg17153794', 'cg17153876', 'cg17153924', 'cg17153935', 'cg17153947', '\n",
      "1000 input query terms found no hit:\t['cg17184392', 'cg17184430', 'cg17184448', 'cg17184458', 'cg17184477', 'cg17184479', 'cg17184491', '\n",
      "1000 input query terms found no hit:\t['cg17215706', 'cg17215726', 'cg17215819', 'cg17215831', 'cg17215843', 'cg17215934', 'cg17215972', '\n",
      "1000 input query terms found no hit:\t['cg17248729', 'cg17248730', 'cg17248752', 'cg17248770', 'cg17248791', 'cg17248847', 'cg17248909', '\n",
      "1000 input query terms found no hit:\t['cg17281126', 'cg17281206', 'cg17281217', 'cg17281297', 'cg17281355', 'cg17281365', 'cg17281378', '\n",
      "1000 input query terms found no hit:\t['cg17311587', 'cg17311626', 'cg17311649', 'cg17311683', 'cg17311725', 'cg17311726', 'cg17311778', '\n",
      "1000 input query terms found no hit:\t['cg17341313', 'cg17341326', 'cg17341343', 'cg17341345', 'cg17341358', 'cg17341361', 'cg17341366', '\n",
      "1000 input query terms found no hit:\t['cg17372426', 'cg17372451', 'cg17372476', 'cg17372494', 'cg17372581', 'cg17372657', 'cg17372738', '\n",
      "1000 input query terms found no hit:\t['cg17405058', 'cg17405117', 'cg17405122', 'cg17405127', 'cg17405139', 'cg17405145', 'cg17405164', '\n",
      "1000 input query terms found no hit:\t['cg17436134', 'cg17436224', 'cg17436293', 'cg17436295', 'cg17436298', 'cg17436327', 'cg17436370', '\n",
      "1000 input query terms found no hit:\t['cg17468553', 'cg17468563', 'cg17468569', 'cg17468616', 'cg17468640', 'cg17468656', 'cg17468665', '\n",
      "1000 input query terms found no hit:\t['cg17499490', 'cg17499657', 'cg17499682', 'cg17499719', 'cg17499729', 'cg17499822', 'cg17499826', '\n",
      "1000 input query terms found no hit:\t['cg17528648', 'cg17528662', 'cg17528700', 'cg17528768', 'cg17528779', 'cg17528813', 'cg17528880', '\n",
      "1000 input query terms found no hit:\t['cg17562963', 'cg17562972', 'cg17563032', 'cg17563034', 'cg17563088', 'cg17563106', 'cg17563146', '\n",
      "1000 input query terms found no hit:\t['cg17597901', 'cg17597911', 'cg17597915', 'cg17598013', 'cg17598128', 'cg17598140', 'cg17598145', '\n",
      "1000 input query terms found no hit:\t['cg17632124', 'cg17632170', 'cg17632214', 'cg17632226', 'cg17632233', 'cg17632243', 'cg17632299', '\n",
      "1000 input query terms found no hit:\t['cg17667113', 'cg17667145', 'cg17667195', 'cg17667217', 'cg17667220', 'cg17667229', 'cg17667294', '\n",
      "1000 input query terms found no hit:\t['cg17703078', 'cg17703079', 'cg17703081', 'cg17703087', 'cg17703135', 'cg17703212', 'cg17703214', '\n",
      "1000 input query terms found no hit:\t['cg17736895', 'cg17736905', 'cg17736920', 'cg17736928', 'cg17736936', 'cg17736940', 'cg17736982', '\n",
      "1000 input query terms found no hit:\t['cg17768253', 'cg17768313', 'cg17768409', 'cg17768439', 'cg17768474', 'cg17768491', 'cg17768502', '\n",
      "1000 input query terms found no hit:\t['cg17804779', 'cg17804825', 'cg17804850', 'cg17804886', 'cg17804911', 'cg17804959', 'cg17804981', '\n",
      "1000 input query terms found no hit:\t['cg17837792', 'cg17837888', 'cg17837894', 'cg17837918', 'cg17837967', 'cg17838000', 'cg17838026', '\n",
      "1000 input query terms found no hit:\t['cg17871993', 'cg17872064', 'cg17872110', 'cg17872128', 'cg17872135', 'cg17872136', 'cg17872261', '\n",
      "1000 input query terms found no hit:\t['cg17907059', 'cg17907120', 'cg17907128', 'cg17907144', 'cg17907200', 'cg17907238', 'cg17907351', '\n",
      "1000 input query terms found no hit:\t['cg17944603', 'cg17944616', 'cg17944645', 'cg17944703', 'cg17944737', 'cg17944741', 'cg17944774', '\n",
      "1000 input query terms found no hit:\t['cg17981058', 'cg17981082', 'cg17981091', 'cg17981101', 'cg17981102', 'cg17981107', 'cg17981236', '\n",
      "1000 input query terms found no hit:\t['cg18015777', 'cg18015787', 'cg18015805', 'cg18015807', 'cg18015809', 'cg18015816', 'cg18015844', '\n",
      "1000 input query terms found no hit:\t['cg18051234', 'cg18051316', 'cg18051346', 'cg18051353', 'cg18051361', 'cg18051433', 'cg18051449', '\n",
      "1000 input query terms found no hit:\t['cg18085869', 'cg18085925', 'cg18085990', 'cg18085998', 'cg18086083', 'cg18086100', 'cg18086113', '\n",
      "1000 input query terms found no hit:\t['cg18117798', 'cg18117827', 'cg18117847', 'cg18117871', 'cg18117895', 'cg18118000', 'cg18118033', '\n",
      "1000 input query terms found no hit:\t['cg18150885', 'cg18150909', 'cg18150958', 'cg18150978', 'cg18151030', 'cg18151041', 'cg18151074', '\n",
      "1000 input query terms found no hit:\t['cg18185980', 'cg18185987', 'cg18186081', 'cg18186235', 'cg18186241', 'cg18186308', 'cg18186315', '\n",
      "1000 input query terms found no hit:\t['cg18222609', 'cg18222653', 'cg18222670', 'cg18222678', 'cg18222759', 'cg18222768', 'cg18222769', '\n",
      "1000 input query terms found no hit:\t['cg18256894', 'cg18256897', 'cg18256949', 'cg18256985', 'cg18257026', 'cg18257031', 'cg18257103', '\n",
      "1000 input query terms found no hit:\t['cg18292394', 'cg18292434', 'cg18292570', 'cg18292664', 'cg18292711', 'cg18292763', 'cg18292904', '\n",
      "1000 input query terms found no hit:\t['cg18329187', 'cg18329216', 'cg18329218', 'cg18329349', 'cg18329352', 'cg18329418', 'cg18329423', '\n",
      "1000 input query terms found no hit:\t['cg18365865', 'cg18365888', 'cg18365932', 'cg18365938', 'cg18365960', 'cg18366108', 'cg18366149', '\n",
      "1000 input query terms found no hit:\t['cg18401658', 'cg18401753', 'cg18401778', 'cg18401934', 'cg18401981', 'cg18402034', 'cg18402101', '\n",
      "1000 input query terms found no hit:\t['cg18437077', 'cg18437099', 'cg18437144', 'cg18437152', 'cg18437193', 'cg18437209', 'cg18437219', '\n",
      "1000 input query terms found no hit:\t['cg18470839', 'cg18470849', 'cg18470852', 'cg18470886', 'cg18470891', 'cg18470904', 'cg18470925', '\n",
      "1000 input query terms found no hit:\t['cg18504653', 'cg18504656', 'cg18504661', 'cg18504691', 'cg18504708', 'cg18504811', 'cg18504826', '\n",
      "1000 input query terms found no hit:\t['cg18541608', 'cg18541609', 'cg18541720', 'cg18541841', 'cg18541880', 'cg18541904', 'cg18542050', '\n",
      "1000 input query terms found no hit:\t['cg18577231', 'cg18577232', 'cg18577239', 'cg18577280', 'cg18577326', 'cg18577456', 'cg18577457', '\n",
      "1000 input query terms found no hit:\t['cg18612805', 'cg18612880', 'cg18612940', 'cg18612955', 'cg18612956', 'cg18613031', 'cg18613092', '\n",
      "1000 input query terms found no hit:\t['cg18649203', 'cg18649228', 'cg18649283', 'cg18649310', 'cg18649319', 'cg18649323', 'cg18649324', '\n",
      "1000 input query terms found no hit:\t['cg18687314', 'cg18687347', 'cg18687468', 'cg18687492', 'cg18687533', 'cg18687588', 'cg18687621', '\n",
      "1000 input query terms found no hit:\t['cg18722861', 'cg18722881', 'cg18722919', 'cg18722934', 'cg18722942', 'cg18722949', 'cg18723031', '\n",
      "1000 input query terms found no hit:\t['cg18757824', 'cg18757828', 'cg18757935', 'cg18757941', 'cg18757962', 'cg18757974', 'cg18758031', '\n",
      "1000 input query terms found no hit:\t['cg18790649', 'cg18790730', 'cg18790741', 'cg18790771', 'cg18790774', 'cg18790787', 'cg18790803', '\n",
      "1000 input query terms found no hit:\t['cg18824560', 'cg18824571', 'cg18824585', 'cg18824596', 'cg18824598', 'cg18824724', 'cg18824840', '\n",
      "1000 input query terms found no hit:\t['cg18861189', 'cg18861197', 'cg18861222', 'cg18861231', 'cg18861250', 'cg18861300', 'cg18861311', '\n",
      "1000 input query terms found no hit:\t['cg18897025', 'cg18897026', 'cg18897041', 'cg18897157', 'cg18897177', 'cg18897206', 'cg18897249', '\n",
      "1000 input query terms found no hit:\t['cg18937774', 'cg18937827', 'cg18937851', 'cg18937875', 'cg18937897', 'cg18937932', 'cg18937934', '\n",
      "1000 input query terms found no hit:\t['cg18976542', 'cg18976563', 'cg18976571', 'cg18976581', 'cg18976649', 'cg18976667', 'cg18976765', '\n",
      "1000 input query terms found no hit:\t['cg19013682', 'cg19013695', 'cg19013742', 'cg19013753', 'cg19013813', 'cg19013966', 'cg19014002', '\n",
      "1000 input query terms found no hit:\t['cg19049165', 'cg19049167', 'cg19049194', 'cg19049213', 'cg19049246', 'cg19049305', 'cg19049316', '\n",
      "1000 input query terms found no hit:\t['cg19088961', 'cg19088962', 'cg19088984', 'cg19089033', 'cg19089045', 'cg19089070', 'cg19089073', '\n",
      "1000 input query terms found no hit:\t['cg19123650', 'cg19123748', 'cg19123749', 'cg19123750', 'cg19123846', 'cg19123863', 'cg19123882', '\n",
      "1000 input query terms found no hit:\t['cg19162106', 'cg19162135', 'cg19162300', 'cg19162333', 'cg19162342', 'cg19162416', 'cg19162470', '\n",
      "1000 input query terms found no hit:\t['cg19202157', 'cg19202213', 'cg19202269', 'cg19202288', 'cg19202292', 'cg19202369', 'cg19202384', '\n",
      "1000 input query terms found no hit:\t['cg19240021', 'cg19240039', 'cg19240040', 'cg19240052', 'cg19240090', 'cg19240103', 'cg19240110', '\n",
      "1000 input query terms found no hit:\t['cg19273253', 'cg19273262', 'cg19273316', 'cg19273347', 'cg19273378', 'cg19273387', 'cg19273422', '\n",
      "1000 input query terms found no hit:\t['cg19310322', 'cg19310340', 'cg19310365', 'cg19310375', 'cg19310420', 'cg19310430', 'cg19310458', '\n",
      "1000 input query terms found no hit:\t['cg19350970', 'cg19350992', 'cg19351000', 'cg19351026', 'cg19351088', 'cg19351107', 'cg19351143', '\n",
      "1000 input query terms found no hit:\t['cg19388273', 'cg19388280', 'cg19388312', 'cg19388332', 'cg19388373', 'cg19388411', 'cg19388446', '\n",
      "1000 input query terms found no hit:\t['cg19422565', 'cg19422575', 'cg19422652', 'cg19422657', 'cg19422661', 'cg19422687', 'cg19422705', '\n",
      "1000 input query terms found no hit:\t['cg19458410', 'cg19458452', 'cg19458485', 'cg19458497', 'cg19458529', 'cg19458593', 'cg19458602', '\n",
      "1000 input query terms found no hit:\t['cg19497021', 'cg19497031', 'cg19497037', 'cg19497071', 'cg19497082', 'cg19497114', 'cg19497127', '\n",
      "1000 input query terms found no hit:\t['cg19529596', 'cg19529599', 'cg19529604', 'cg19529621', 'cg19529645', 'cg19529675', 'cg19529696', '\n",
      "1000 input query terms found no hit:\t['cg19564603', 'cg19564615', 'cg19564630', 'cg19564637', 'cg19564648', 'cg19564816', 'cg19564867', '\n",
      "1000 input query terms found no hit:\t['cg19599776', 'cg19599782', 'cg19599803', 'cg19599827', 'cg19599856', 'cg19599862', 'cg19599951', '\n",
      "1000 input query terms found no hit:\t['cg19636366', 'cg19636384', 'cg19636411', 'cg19636492', 'cg19636519', 'cg19636550', 'cg19636625', '\n",
      "1000 input query terms found no hit:\t['cg19673233', 'cg19673237', 'cg19673302', 'cg19673329', 'cg19673337', 'cg19673350', 'cg19673376', '\n",
      "1000 input query terms found no hit:\t['cg19707450', 'cg19707453', 'cg19707454', 'cg19707503', 'cg19707610', 'cg19707638', 'cg19707653', '\n",
      "1000 input query terms found no hit:\t['cg19741727', 'cg19741767', 'cg19741777', 'cg19741792', 'cg19741821', 'cg19741838', 'cg19741868', '\n",
      "1000 input query terms found no hit:\t['cg19773671', 'cg19773937', 'cg19773942', 'cg19774031', 'cg19774146', 'cg19774218', 'cg19774236', '\n",
      "1000 input query terms found no hit:\t['cg19808643', 'cg19808663', 'cg19808737', 'cg19808825', 'cg19808829', 'cg19808841', 'cg19808844', '\n",
      "1000 input query terms found no hit:\t['cg19843760', 'cg19843891', 'cg19843939', 'cg19843977', 'cg19843991', 'cg19844051', 'cg19844101', '\n",
      "1000 input query terms found no hit:\t['cg19877840', 'cg19877843', 'cg19877921', 'cg19878017', 'cg19878076', 'cg19878084', 'cg19878120', '\n",
      "1000 input query terms found no hit:\t['cg19913496', 'cg19913537', 'cg19913551', 'cg19913557', 'cg19913559', 'cg19913563', 'cg19913580', '\n",
      "1000 input query terms found no hit:\t['cg19946725', 'cg19946734', 'cg19946742', 'cg19946916', 'cg19946956', 'cg19946982', 'cg19947009', '\n",
      "1000 input query terms found no hit:\t['cg19982868', 'cg19982922', 'cg19982958', 'cg19982971', 'cg19983027', 'cg19983103', 'cg19983118', '\n",
      "1000 input query terms found no hit:\t['cg20019019', 'cg20019062', 'cg20019086', 'cg20019107', 'cg20019163', 'cg20019168', 'cg20019202', '\n",
      "1000 input query terms found no hit:\t['cg20056298', 'cg20056324', 'cg20056341', 'cg20056362', 'cg20056542', 'cg20056593', 'cg20056602', '\n",
      "1000 input query terms found no hit:\t['cg20092892', 'cg20092936', 'cg20092942', 'cg20092944', 'cg20092999', 'cg20093013', 'cg20093020', '\n",
      "1000 input query terms found no hit:\t['cg20130018', 'cg20130062', 'cg20130085', 'cg20130097', 'cg20130098', 'cg20130138', 'cg20130344', '\n",
      "1000 input query terms found no hit:\t['cg20167197', 'cg20167216', 'cg20167264', 'cg20167290', 'cg20167293', 'cg20167366', 'cg20167446', '\n",
      "1000 input query terms found no hit:\t['cg20207973', 'cg20207982', 'cg20207989', 'cg20208009', 'cg20208021', 'cg20208052', 'cg20208061', '\n",
      "1000 input query terms found no hit:\t['cg20245328', 'cg20245361', 'cg20245362', 'cg20245379', 'cg20245439', 'cg20245444', 'cg20245503', '\n",
      "1000 input query terms found no hit:\t['cg20281075', 'cg20281082', 'cg20281105', 'cg20281117', 'cg20281264', 'cg20281309', 'cg20281375', '\n",
      "1000 input query terms found no hit:\t['cg20317180', 'cg20317252', 'cg20317272', 'cg20317382', 'cg20317417', 'cg20317418', 'cg20317437', '\n",
      "1000 input query terms found no hit:\t['cg20353692', 'cg20353750', 'cg20353780', 'cg20353857', 'cg20353881', 'cg20353884', 'cg20353935', '\n",
      "1000 input query terms found no hit:\t['cg20391652', 'cg20391668', 'cg20391733', 'cg20391764', 'cg20391784', 'cg20391793', 'cg20391833', '\n",
      "1000 input query terms found no hit:\t['cg20428033', 'cg20428068', 'cg20428112', 'cg20428120', 'cg20428133', 'cg20428144', 'cg20428153', '\n",
      "1000 input query terms found no hit:\t['cg20463539', 'cg20463564', 'cg20463590', 'cg20463715', 'cg20463741', 'cg20463750', 'cg20463752', '\n",
      "1000 input query terms found no hit:\t['cg20501355', 'cg20501434', 'cg20501491', 'cg20501512', 'cg20501518', 'cg20501527', 'cg20501697', '\n",
      "1000 input query terms found no hit:\t['cg20540566', 'cg20540582', 'cg20540600', 'cg20540608', 'cg20540629', 'cg20540672', 'cg20540683', '\n",
      "1000 input query terms found no hit:\t['cg20576243', 'cg20576264', 'cg20576272', 'cg20576319', 'cg20576322', 'cg20576328', 'cg20576349', '\n",
      "1000 input query terms found no hit:\t['cg20615843', 'cg20615848', 'cg20615879', 'cg20615910', 'cg20615951', 'cg20616045', 'cg20616060', '\n",
      "1000 input query terms found no hit:\t['cg20652630', 'cg20652636', 'cg20652640', 'cg20652681', 'cg20652749', 'cg20652759', 'cg20652790', '\n",
      "1000 input query terms found no hit:\t['cg20686584', 'cg20686658', 'cg20686679', 'cg20686720', 'cg20686742', 'cg20686755', 'cg20686777', '\n",
      "1000 input query terms found no hit:\t['cg20717875', 'cg20717908', 'cg20717932', 'cg20717983', 'cg20718051', 'cg20718134', 'cg20718169', '\n",
      "1000 input query terms found no hit:\t['cg20754867', 'cg20754881', 'cg20754964', 'cg20754979', 'cg20754980', 'cg20754983', 'cg20755040', '\n",
      "1000 input query terms found no hit:\t['cg20793089', 'cg20793144', 'cg20793174', 'cg20793193', 'cg20793303', 'cg20793307', 'cg20793341', '\n",
      "1000 input query terms found no hit:\t['cg20829148', 'cg20829176', 'cg20829179', 'cg20829193', 'cg20829211', 'cg20829233', 'cg20829256', '\n",
      "1000 input query terms found no hit:\t['cg20870340', 'cg20870360', 'cg20870362', 'cg20870368', 'cg20870372', 'cg20870394', 'cg20870512', '\n",
      "1000 input query terms found no hit:\t['cg20909354', 'cg20909355', 'cg20909380', 'cg20909399', 'cg20909418', 'cg20909531', 'cg20909579', '\n",
      "1000 input query terms found no hit:\t['cg20945376', 'cg20945455', 'cg20945463', 'cg20945465', 'cg20945472', 'cg20945491', 'cg20945509', '\n",
      "1000 input query terms found no hit:\t['cg20979815', 'cg20979823', 'cg20979852', 'cg20979896', 'cg20979921', 'cg20979926', 'cg20979979', '\n",
      "1000 input query terms found no hit:\t['cg21015181', 'cg21015266', 'cg21015393', 'cg21015470', 'cg21015471', 'cg21015554', 'cg21015582', '\n",
      "1000 input query terms found no hit:\t['cg21049762', 'cg21049809', 'cg21049825', 'cg21049840', 'cg21049843', 'cg21049958', 'cg21049965', '\n",
      "1000 input query terms found no hit:\t['cg21088542', 'cg21088646', 'cg21088686', 'cg21088715', 'cg21088783', 'cg21088896', 'cg21088983', '\n",
      "1000 input query terms found no hit:\t['cg21126404', 'cg21126453', 'cg21126472', 'cg21126487', 'cg21126506', 'cg21126573', 'cg21126583', '\n",
      "1000 input query terms found no hit:\t['cg21160290', 'cg21160328', 'cg21160344', 'cg21160357', 'cg21160363', 'cg21160372', 'cg21160379', '\n",
      "1000 input query terms found no hit:\t['cg21192344', 'cg21192370', 'cg21192373', 'cg21192376', 'cg21192452', 'cg21192468', 'cg21192579', '\n",
      "1000 input query terms found no hit:\t['cg21223094', 'cg21223117', 'cg21223135', 'cg21223159', 'cg21223191', 'cg21223206', 'cg21223214', '\n",
      "1000 input query terms found no hit:\t['cg21254855', 'cg21254856', 'cg21254883', 'cg21254912', 'cg21254939', 'cg21255097', 'cg21255128', '\n",
      "1000 input query terms found no hit:\t['cg21294654', 'cg21294655', 'cg21294714', 'cg21294812', 'cg21294847', 'cg21294861', 'cg21294894', '\n",
      "1000 input query terms found no hit:\t['cg21334129', 'cg21334144', 'cg21334191', 'cg21334198', 'cg21334395', 'cg21334405', 'cg21334449', '\n",
      "1000 input query terms found no hit:\t['cg21375878', 'cg21375895', 'cg21375928', 'cg21376004', 'cg21376058', 'cg21376090', 'cg21376109', '\n",
      "1000 input query terms found no hit:\t['cg21414268', 'cg21414336', 'cg21414363', 'cg21414369', 'cg21414411', 'cg21414424', 'cg21414538', '\n",
      "1000 input query terms found no hit:\t['cg21452443', 'cg21452447', 'cg21452471', 'cg21452476', 'cg21452511', 'cg21452538', 'cg21452601', '\n",
      "1000 input query terms found no hit:\t['cg21488132', 'cg21488156', 'cg21488279', 'cg21488289', 'cg21488396', 'cg21488398', 'cg21488474', '\n",
      "1000 input query terms found no hit:\t['cg21523564', 'cg21523574', 'cg21523601', 'cg21523613', 'cg21523680', 'cg21523688', 'cg21523701', '\n",
      "1000 input query terms found no hit:\t['cg21558409', 'cg21558423', 'cg21558485', 'cg21558490', 'cg21558508', 'cg21558509', 'cg21558522', '\n",
      "1000 input query terms found no hit:\t['cg21593148', 'cg21593149', 'cg21593159', 'cg21593161', 'cg21593175', 'cg21593211', 'cg21593216', '\n",
      "1000 input query terms found no hit:\t['cg21628055', 'cg21628164', 'cg21628200', 'cg21628297', 'cg21628390', 'cg21628404', 'cg21628485', '\n",
      "1000 input query terms found no hit:\t['cg21665242', 'cg21665288', 'cg21665294', 'cg21665350', 'cg21665431', 'cg21665448', 'cg21665502', '\n",
      "1000 input query terms found no hit:\t['cg21705513', 'cg21705584', 'cg21705600', 'cg21705625', 'cg21705669', 'cg21705712', 'cg21705844', '\n",
      "1000 input query terms found no hit:\t['cg21746339', 'cg21746387', 'cg21746409', 'cg21746425', 'cg21746455', 'cg21746459', 'cg21746472', '\n",
      "1000 input query terms found no hit:\t['cg21783672', 'cg21783686', 'cg21783716', 'cg21783725', 'cg21783729', 'cg21783749', 'cg21783852', '\n",
      "1000 input query terms found no hit:\t['cg21820609', 'cg21820610', 'cg21820652', 'cg21820656', 'cg21820677', 'cg21820708', 'cg21820712', '\n",
      "1000 input query terms found no hit:\t['cg21855329', 'cg21855360', 'cg21855392', 'cg21855450', 'cg21855467', 'cg21855587', 'cg21855617', '\n",
      "1000 input query terms found no hit:\t['cg21891922', 'cg21891967', 'cg21892085', 'cg21892111', 'cg21892150', 'cg21892242', 'cg21892257', '\n",
      "1000 input query terms found no hit:\t['cg21927021', 'cg21927055', 'cg21927086', 'cg21927108', 'cg21927122', 'cg21927177', 'cg21927242', '\n",
      "1000 input query terms found no hit:\t['cg21963392', 'cg21963406', 'cg21963436', 'cg21963535', 'cg21963544', 'cg21963573', 'cg21963583', '\n",
      "1000 input query terms found no hit:\t['cg22005100', 'cg22005145', 'cg22005150', 'cg22005204', 'cg22005304', 'cg22005336', 'cg22005346', '\n",
      "1000 input query terms found no hit:\t['cg22040162', 'cg22040202', 'cg22040239', 'cg22040274', 'cg22040301', 'cg22040303', 'cg22040335', '\n",
      "1000 input query terms found no hit:\t['cg22076702', 'cg22076755', 'cg22076757', 'cg22076764', 'cg22076769', 'cg22076873', 'cg22076912', '\n",
      "1000 input query terms found no hit:\t['cg22115502', 'cg22115549', 'cg22115561', 'cg22115596', 'cg22115597', 'cg22115697', 'cg22115706', '\n",
      "1000 input query terms found no hit:\t['cg22152229', 'cg22152253', 'cg22152268', 'cg22152279', 'cg22152328', 'cg22152397', 'cg22152407', '\n",
      "1000 input query terms found no hit:\t['cg22189952', 'cg22189991', 'cg22190023', 'cg22190028', 'cg22190077', 'cg22190090', 'cg22190114', '\n",
      "1000 input query terms found no hit:\t['cg22228982', 'cg22228988', 'cg22229016', 'cg22229034', 'cg22229039', 'cg22229043', 'cg22229076', '\n",
      "1000 input query terms found no hit:\t['cg22265441', 'cg22265452', 'cg22265458', 'cg22265501', 'cg22265539', 'cg22265544', 'cg22265605', '\n",
      "1000 input query terms found no hit:\t['cg22305782', 'cg22305797', 'cg22305826', 'cg22305850', 'cg22305921', 'cg22305971', 'cg22306009', '\n",
      "1000 input query terms found no hit:\t['cg22344413', 'cg22344419', 'cg22344533', 'cg22344544', 'cg22344549', 'cg22344564', 'cg22344578', '\n",
      "1000 input query terms found no hit:\t['cg22380472', 'cg22380476', 'cg22380489', 'cg22380508', 'cg22380533', 'cg22380613', 'cg22380619', '\n",
      "1000 input query terms found no hit:\t['cg22419864', 'cg22419886', 'cg22419888', 'cg22419965', 'cg22420044', 'cg22420249', 'cg22420297', '\n",
      "1000 input query terms found no hit:\t['cg22460443', 'cg22460512', 'cg22460569', 'cg22460586', 'cg22460590', 'cg22460601', 'cg22460642', '\n",
      "1000 input query terms found no hit:\t['cg22498251', 'cg22498365', 'cg22498396', 'cg22498404', 'cg22498452', 'cg22498453', 'cg22498467', '\n",
      "1000 input query terms found no hit:\t['cg22533677', 'cg22533681', 'cg22533683', 'cg22533689', 'cg22533717', 'cg22533724', 'cg22533762', '\n",
      "1000 input query terms found no hit:\t['cg22570053', 'cg22570082', 'cg22570104', 'cg22570122', 'cg22570213', 'cg22570250', 'cg22570288', '\n",
      "1000 input query terms found no hit:\t['cg22608646', 'cg22608655', 'cg22608657', 'cg22608691', 'cg22608764', 'cg22608800', 'cg22608845', '\n",
      "1000 input query terms found no hit:\t['cg22646227', 'cg22646311', 'cg22646324', 'cg22646408', 'cg22646443', 'cg22646528', 'cg22646570', '\n",
      "1000 input query terms found no hit:\t['cg22682201', 'cg22682213', 'cg22682215', 'cg22682254', 'cg22682296', 'cg22682297', 'cg22682299', '\n",
      "1000 input query terms found no hit:\t['cg22715220', 'cg22715246', 'cg22715321', 'cg22715398', 'cg22715442', 'cg22715479', 'cg22715531', '\n",
      "1000 input query terms found no hit:\t['cg22752586', 'cg22752588', 'cg22752602', 'cg22752623', 'cg22752827', 'cg22752869', 'cg22752924', '\n",
      "1000 input query terms found no hit:\t['cg22790973', 'cg22790986', 'cg22791030', 'cg22791045', 'cg22791053', 'cg22791155', 'cg22791163', '\n",
      "1000 input query terms found no hit:\t['cg22823987', 'cg22824002', 'cg22824060', 'cg22824265', 'cg22824274', 'cg22824291', 'cg22824293', '\n",
      "1000 input query terms found no hit:\t['cg22859451', 'cg22859492', 'cg22859541', 'cg22859585', 'cg22859597', 'cg22859658', 'cg22859679', '\n",
      "1000 input query terms found no hit:\t['cg22891588', 'cg22891595', 'cg22891600', 'cg22891601', 'cg22891619', 'cg22891632', 'cg22891637', '\n",
      "1000 input query terms found no hit:\t['cg22930505', 'cg22930549', 'cg22930587', 'cg22930588', 'cg22930612', 'cg22930650', 'cg22930677', '\n",
      "1000 input query terms found no hit:\t['cg22965943', 'cg22965956', 'cg22966003', 'cg22966026', 'cg22966060', 'cg22966065', 'cg22966109', '\n",
      "1000 input query terms found no hit:\t['cg23002415', 'cg23002431', 'cg23002439', 'cg23002562', 'cg23002586', 'cg23002590', 'cg23002595', '\n",
      "1000 input query terms found no hit:\t['cg23036136', 'cg23036148', 'cg23036166', 'cg23036171', 'cg23036194', 'cg23036204', 'cg23036213', '\n",
      "1000 input query terms found no hit:\t['cg23069305', 'cg23069433', 'cg23069491', 'cg23069551', 'cg23069648', 'cg23069677', 'cg23069684', '\n",
      "1000 input query terms found no hit:\t['cg23104439', 'cg23104481', 'cg23104520', 'cg23104539', 'cg23104543', 'cg23104549', 'cg23104590', '\n",
      "1000 input query terms found no hit:\t['cg23143902', 'cg23143978', 'cg23143991', 'cg23144046', 'cg23144115', 'cg23144123', 'cg23144349', '\n",
      "1000 input query terms found no hit:\t['cg23181554', 'cg23181573', 'cg23181580', 'cg23181590', 'cg23181591', 'cg23181627', 'cg23181703', '\n",
      "1000 input query terms found no hit:\t['cg23214565', 'cg23214606', 'cg23214628', 'cg23214755', 'cg23214760', 'cg23214764', 'cg23214787', '\n",
      "1000 input query terms found no hit:\t['cg23247475', 'cg23247476', 'cg23247504', 'cg23247524', 'cg23247595', 'cg23247602', 'cg23247616', '\n",
      "1000 input query terms found no hit:\t['cg23282179', 'cg23282202', 'cg23282276', 'cg23282363', 'cg23282417', 'cg23282441', 'cg23282442', '\n",
      "1000 input query terms found no hit:\t['cg23320522', 'cg23320550', 'cg23320649', 'cg23320654', 'cg23320693', 'cg23320787', 'cg23320834', '\n",
      "1000 input query terms found no hit:\t['cg23355472', 'cg23355492', 'cg23355504', 'cg23355537', 'cg23355546', 'cg23355627', 'cg23355637', '\n",
      "1000 input query terms found no hit:\t['cg23391021', 'cg23391041', 'cg23391045', 'cg23391086', 'cg23391090', 'cg23391098', 'cg23391107', '\n",
      "1000 input query terms found no hit:\t['cg23427255', 'cg23427269', 'cg23427282', 'cg23427348', 'cg23427362', 'cg23427394', 'cg23427404', '\n",
      "1000 input query terms found no hit:\t['cg23463531', 'cg23463533', 'cg23463608', 'cg23463631', 'cg23463642', 'cg23463680', 'cg23463700', '\n",
      "1000 input query terms found no hit:\t['cg23498494', 'cg23498518', 'cg23498547', 'cg23498628', 'cg23498739', 'cg23498748', 'cg23498771', '\n",
      "1000 input query terms found no hit:\t['cg23531317', 'cg23531340', 'cg23531352', 'cg23531459', 'cg23531468', 'cg23531488', 'cg23531521', '\n",
      "1000 input query terms found no hit:\t['cg23568910', 'cg23568912', 'cg23568913', 'cg23568976', 'cg23568998', 'cg23569004', 'cg23569034', '\n",
      "1000 input query terms found no hit:\t['cg23606718', 'cg23606722', 'cg23606751', 'cg23606758', 'cg23606775', 'cg23606809', 'cg23606861', '\n",
      "1000 input query terms found no hit:\t['cg23640500', 'cg23640575', 'cg23640587', 'cg23640615', 'cg23640625', 'cg23640664', 'cg23640703', '\n",
      "1000 input query terms found no hit:\t['cg23673509', 'cg23673553', 'cg23673588', 'cg23673590', 'cg23673620', 'cg23673732', 'cg23673761', '\n",
      "1000 input query terms found no hit:\t['cg23704085', 'cg23704092', 'cg23704097', 'cg23704107', 'cg23704111', 'cg23704137', 'cg23704141', '\n",
      "1000 input query terms found no hit:\t['cg23738830', 'cg23738831', 'cg23738833', 'cg23738842', 'cg23738872', 'cg23738973', 'cg23739027', '\n",
      "1000 input query terms found no hit:\t['cg23780614', 'cg23780635', 'cg23780713', 'cg23780731', 'cg23780797', 'cg23780875', 'cg23780901', '\n",
      "1000 input query terms found no hit:\t['cg23818888', 'cg23818910', 'cg23818921', 'cg23818931', 'cg23818937', 'cg23818978', 'cg23819016', '\n",
      "1000 input query terms found no hit:\t['cg23854009', 'cg23854042', 'cg23854087', 'cg23854103', 'cg23854195', 'cg23854219', 'cg23854246', '\n",
      "1000 input query terms found no hit:\t['cg23889914', 'cg23889940', 'cg23889967', 'cg23889993', 'cg23890008', 'cg23890044', 'cg23890059', '\n",
      "1000 input query terms found no hit:\t['cg23924111', 'cg23924137', 'cg23924222', 'cg23924229', 'cg23924241', 'cg23924282', 'cg23924306', '\n",
      "1000 input query terms found no hit:\t['cg23957915', 'cg23957920', 'cg23957929', 'cg23958008', 'cg23958023', 'cg23958118', 'cg23958151', '\n",
      "1000 input query terms found no hit:\t['cg23991947', 'cg23991989', 'cg23992003', 'cg23992051', 'cg23992123', 'cg23992130', 'cg23992143', '\n",
      "1000 input query terms found no hit:\t['cg24025566', 'cg24025567', 'cg24025627', 'cg24025644', 'cg24025651', 'cg24025716', 'cg24025721', '\n",
      "1000 input query terms found no hit:\t['cg24059366', 'cg24059404', 'cg24059418', 'cg24059612', 'cg24059623', 'cg24059677', 'cg24059708', '\n",
      "1000 input query terms found no hit:\t['cg24094775', 'cg24094804', 'cg24094848', 'cg24094850', 'cg24094897', 'cg24095088', 'cg24095120', '\n",
      "1000 input query terms found no hit:\t['cg24129820', 'cg24129873', 'cg24129904', 'cg24129914', 'cg24129923', 'cg24129947', 'cg24129977', '\n",
      "1000 input query terms found no hit:\t['cg24163418', 'cg24163448', 'cg24163557', 'cg24163563', 'cg24163568', 'cg24163575', 'cg24163613', '\n",
      "1000 input query terms found no hit:\t['cg24199687', 'cg24199724', 'cg24199766', 'cg24199834', 'cg24199987', 'cg24200022', 'cg24200059', '\n",
      "1000 input query terms found no hit:\t['cg24239922', 'cg24239961', 'cg24239992', 'cg24239999', 'cg24240077', 'cg24240169', 'cg24240172', '\n",
      "1000 input query terms found no hit:\t['cg24279744', 'cg24279750', 'cg24279836', 'cg24279837', 'cg24279897', 'cg24279914', 'cg24279934', '\n",
      "1000 input query terms found no hit:\t['cg24315703', 'cg24315709', 'cg24315710', 'cg24315757', 'cg24315770', 'cg24315782', 'cg24315815', '\n",
      "1000 input query terms found no hit:\t['cg24349555', 'cg24349573', 'cg24349610', 'cg24349631', 'cg24349665', 'cg24349668', 'cg24349745', '\n",
      "1000 input query terms found no hit:\t['cg24384676', 'cg24384713', 'cg24384816', 'cg24384817', 'cg24384879', 'cg24384882', 'cg24384918', '\n",
      "1000 input query terms found no hit:\t['cg24418956', 'cg24419040', 'cg24419071', 'cg24419094', 'cg24419099', 'cg24419101', 'cg24419108', '\n",
      "1000 input query terms found no hit:\t['cg24454374', 'cg24454435', 'cg24454440', 'cg24454448', 'cg24454553', 'cg24454579', 'cg24454673', '\n",
      "1000 input query terms found no hit:\t['cg24489626', 'cg24489634', 'cg24489686', 'cg24489689', 'cg24489697', 'cg24489759', 'cg24489774', '\n",
      "1000 input query terms found no hit:\t['cg24524435', 'cg24524451', 'cg24524470', 'cg24524485', 'cg24524522', 'cg24524538', 'cg24524539', '\n",
      "1000 input query terms found no hit:\t['cg24557378', 'cg24557443', 'cg24557470', 'cg24557546', 'cg24557547', 'cg24557620', 'cg24557627', '\n",
      "1000 input query terms found no hit:\t['cg24595152', 'cg24595208', 'cg24595261', 'cg24595270', 'cg24595310', 'cg24595440', 'cg24595510', '\n",
      "1000 input query terms found no hit:\t['cg24630733', 'cg24630778', 'cg24630782', 'cg24630825', 'cg24630826', 'cg24630878', 'cg24630943', '\n",
      "1000 input query terms found no hit:\t['cg24666196', 'cg24666248', 'cg24666273', 'cg24666276', 'cg24666355', 'cg24666389', 'cg24666521', '\n",
      "1000 input query terms found no hit:\t['cg24698564', 'cg24698622', 'cg24698655', 'cg24698727', 'cg24698761', 'cg24698774', 'cg24698780', '\n",
      "1000 input query terms found no hit:\t['cg24729839', 'cg24729879', 'cg24729928', 'cg24729955', 'cg24729968', 'cg24729983', 'cg24729988', '\n",
      "1000 input query terms found no hit:\t['cg24765648', 'cg24765649', 'cg24765658', 'cg24765668', 'cg24765673', 'cg24765723', 'cg24765748', '\n",
      "1000 input query terms found no hit:\t['cg24799863', 'cg24799894', 'cg24799909', 'cg24799921', 'cg24799932', 'cg24799947', 'cg24800032', '\n",
      "1000 input query terms found no hit:\t['cg24835051', 'cg24835149', 'cg24835153', 'cg24835159', 'cg24835190', 'cg24835232', 'cg24835233', '\n",
      "1000 input query terms found no hit:\t['cg24867820', 'cg24867946', 'cg24867981', 'cg24868015', 'cg24868028', 'cg24868047', 'cg24868054', '\n",
      "1000 input query terms found no hit:\t['cg24902250', 'cg24902269', 'cg24902278', 'cg24902292', 'cg24902339', 'cg24902362', 'cg24902375', '\n",
      "1000 input query terms found no hit:\t['cg24937706', 'cg24937727', 'cg24937735', 'cg24937747', 'cg24937754', 'cg24937768', 'cg24937796', '\n",
      "1000 input query terms found no hit:\t['cg24975986', 'cg24976037', 'cg24976080', 'cg24976089', 'cg24976104', 'cg24976111', 'cg24976131', '\n",
      "1000 input query terms found no hit:\t['cg25011650', 'cg25011652', 'cg25011666', 'cg25011720', 'cg25011749', 'cg25011763', 'cg25011809', '\n",
      "1000 input query terms found no hit:\t['cg25043855', 'cg25043962', 'cg25043986', 'cg25043996', 'cg25044026', 'cg25044100', 'cg25044130', '\n",
      "1000 input query terms found no hit:\t['cg25081796', 'cg25081845', 'cg25081912', 'cg25081977', 'cg25081982', 'cg25081985', 'cg25082000', '\n",
      "1000 input query terms found no hit:\t['cg25118159', 'cg25118167', 'cg25118192', 'cg25118265', 'cg25118283', 'cg25118469', 'cg25118504', '\n",
      "1000 input query terms found no hit:\t['cg25153148', 'cg25153163', 'cg25153196', 'cg25153204', 'cg25153205', 'cg25153233', 'cg25153266', '\n",
      "1000 input query terms found no hit:\t['cg25192651', 'cg25192660', 'cg25192717', 'cg25192768', 'cg25192855', 'cg25192880', 'cg25192916', '\n",
      "1000 input query terms found no hit:\t['cg25226997', 'cg25227066', 'cg25227088', 'cg25227148', 'cg25227163', 'cg25227211', 'cg25227227', '\n",
      "1000 input query terms found no hit:\t['cg25265767', 'cg25265769', 'cg25265790', 'cg25265805', 'cg25265824', 'cg25265929', 'cg25265930', '\n",
      "1000 input query terms found no hit:\t['cg25304100', 'cg25304107', 'cg25304128', 'cg25304129', 'cg25304134', 'cg25304136', 'cg25304146', '\n",
      "1000 input query terms found no hit:\t['cg25338577', 'cg25338581', 'cg25338587', 'cg25338636', 'cg25338707', 'cg25338766', 'cg25338835', '\n",
      "1000 input query terms found no hit:\t['cg25371385', 'cg25371434', 'cg25371449', 'cg25371487', 'cg25371503', 'cg25371538', 'cg25371549', '\n",
      "1000 input query terms found no hit:\t['cg25406779', 'cg25406865', 'cg25406872', 'cg25406923', 'cg25406989', 'cg25407044', 'cg25407053', '\n",
      "1000 input query terms found no hit:\t['cg25441577', 'cg25441591', 'cg25441629', 'cg25441645', 'cg25441661', 'cg25441704', 'cg25441737', '\n",
      "1000 input query terms found no hit:\t['cg25477908', 'cg25477928', 'cg25477957', 'cg25478037', 'cg25478109', 'cg25478112', 'cg25478113', '\n",
      "1000 input query terms found no hit:\t['cg25512683', 'cg25512738', 'cg25512791', 'cg25512848', 'cg25512852', 'cg25512894', 'cg25512903', '\n",
      "1000 input query terms found no hit:\t['cg25551273', 'cg25551287', 'cg25551358', 'cg25551362', 'cg25551529', 'cg25551531', 'cg25551545', '\n",
      "1000 input query terms found no hit:\t['cg25586087', 'cg25586160', 'cg25586209', 'cg25586244', 'cg25586273', 'cg25586299', 'cg25586305', '\n",
      "1000 input query terms found no hit:\t['cg25618981', 'cg25619052', 'cg25619106', 'cg25619136', 'cg25619145', 'cg25619153', 'cg25619157', '\n",
      "1000 input query terms found no hit:\t['cg25654697', 'cg25654705', 'cg25654713', 'cg25654719', 'cg25654768', 'cg25654774', 'cg25654904', '\n",
      "1000 input query terms found no hit:\t['cg25692633', 'cg25692644', 'cg25692669', 'cg25692709', 'cg25692732', 'cg25692739', 'cg25692785', '\n",
      "1000 input query terms found no hit:\t['cg25728994', 'cg25729002', 'cg25729041', 'cg25729060', 'cg25729068', 'cg25729094', 'cg25729132', '\n",
      "1000 input query terms found no hit:\t['cg25762456', 'cg25762522', 'cg25762580', 'cg25762612', 'cg25762654', 'cg25762658', 'cg25762679', '\n",
      "1000 input query terms found no hit:\t['cg25800356', 'cg25800379', 'cg25800472', 'cg25800500', 'cg25800511', 'cg25800527', 'cg25800638', '\n",
      "1000 input query terms found no hit:\t['cg25838080', 'cg25838091', 'cg25838125', 'cg25838150', 'cg25838204', 'cg25838214', 'cg25838230', '\n",
      "1000 input query terms found no hit:\t['cg25879477', 'cg25879514', 'cg25879553', 'cg25879705', 'cg25879745', 'cg25879778', 'cg25879786', '\n",
      "1000 input query terms found no hit:\t['cg25919123', 'cg25919137', 'cg25919177', 'cg25919202', 'cg25919221', 'cg25919345', 'cg25919362', '\n",
      "1000 input query terms found no hit:\t['cg25952447', 'cg25952449', 'cg25952496', 'cg25952510', 'cg25952512', 'cg25952543', 'cg25952573', '\n",
      "1000 input query terms found no hit:\t['cg25987804', 'cg25987814', 'cg25987842', 'cg25987871', 'cg25987923', 'cg25987936', 'cg25987983', '\n",
      "1000 input query terms found no hit:\t['cg26025224', 'cg26025333', 'cg26025358', 'cg26025442', 'cg26025487', 'cg26025505', 'cg26025506', '\n",
      "1000 input query terms found no hit:\t['cg26063037', 'cg26063087', 'cg26063106', 'cg26063133', 'cg26063205', 'cg26063534', 'cg26063563', '\n",
      "1000 input query terms found no hit:\t['cg26098257', 'cg26098311', 'cg26098312', 'cg26098330', 'cg26098386', 'cg26098399', 'cg26098401', '\n",
      "1000 input query terms found no hit:\t['cg26131115', 'cg26131194', 'cg26131246', 'cg26131286', 'cg26131292', 'cg26131306', 'cg26131315', '\n",
      "1000 input query terms found no hit:\t['cg26164151', 'cg26164155', 'cg26164164', 'cg26164184', 'cg26164234', 'cg26164240', 'cg26164269', '\n",
      "1000 input query terms found no hit:\t['cg26200971', 'cg26200974', 'cg26200976', 'cg26200980', 'cg26201011', 'cg26201051', 'cg26201109', '\n",
      "1000 input query terms found no hit:\t['cg26234081', 'cg26234091', 'cg26234137', 'cg26234183', 'cg26234218', 'cg26234223', 'cg26234231', '\n",
      "1000 input query terms found no hit:\t['cg26269109', 'cg26269162', 'cg26269170', 'cg26269171', 'cg26269196', 'cg26269269', 'cg26269286', '\n",
      "1000 input query terms found no hit:\t['cg26302094', 'cg26302103', 'cg26302157', 'cg26302178', 'cg26302190', 'cg26302230', 'cg26302238', '\n",
      "1000 input query terms found no hit:\t['cg26335130', 'cg26335152', 'cg26335251', 'cg26335281', 'cg26335291', 'cg26335299', 'cg26335377', '\n",
      "1000 input query terms found no hit:\t['cg26366663', 'cg26366735', 'cg26366822', 'cg26366833', 'cg26366925', 'cg26366958', 'cg26367013', '\n",
      "1000 input query terms found no hit:\t['cg26402801', 'cg26402811', 'cg26402828', 'cg26402901', 'cg26403043', 'cg26403045', 'cg26403151', '\n",
      "1000 input query terms found no hit:\t['cg26437703', 'cg26437825', 'cg26437826', 'cg26437841', 'cg26437842', 'cg26437845', 'cg26437866', '\n",
      "1000 input query terms found no hit:\t['cg26471781', 'cg26471790', 'cg26471832', 'cg26471866', 'cg26471871', 'cg26471897', 'cg26471982', '\n",
      "1000 input query terms found no hit:\t['cg26506507', 'cg26506524', 'cg26506555', 'cg26506680', 'cg26506725', 'cg26506751', 'cg26506778', '\n",
      "1000 input query terms found no hit:\t['cg26540167', 'cg26540191', 'cg26540230', 'cg26540253', 'cg26540302', 'cg26540315', 'cg26540367', '\n",
      "1000 input query terms found no hit:\t['cg26572677', 'cg26572710', 'cg26572740', 'cg26572770', 'cg26572776', 'cg26572811', 'cg26572887', '\n",
      "1000 input query terms found no hit:\t['cg26610812', 'cg26610868', 'cg26610938', 'cg26611067', 'cg26611070', 'cg26611184', 'cg26611278', '\n",
      "1000 input query terms found no hit:\t['cg26646980', 'cg26647036', 'cg26647111', 'cg26647135', 'cg26647139', 'cg26647197', 'cg26647200', '\n",
      "1000 input query terms found no hit:\t['cg26678479', 'cg26678516', 'cg26678519', 'cg26678575', 'cg26678609', 'cg26678641', 'cg26678648', '\n",
      "1000 input query terms found no hit:\t['cg26710996', 'cg26711011', 'cg26711070', 'cg26711073', 'cg26711145', 'cg26711180', 'cg26711184', '\n",
      "1000 input query terms found no hit:\t['cg26746511', 'cg26746527', 'cg26746533', 'cg26746540', 'cg26746560', 'cg26746573', 'cg26746659', '\n",
      "1000 input query terms found no hit:\t['cg26783691', 'cg26783727', 'cg26783833', 'cg26783952', 'cg26783962', 'cg26784000', 'cg26784011', '\n",
      "1000 input query terms found no hit:\t['cg26816426', 'cg26816432', 'cg26816438', 'cg26816476', 'cg26816491', 'cg26816544', 'cg26816552', '\n",
      "1000 input query terms found no hit:\t['cg26849077', 'cg26849111', 'cg26849212', 'cg26849238', 'cg26849330', 'cg26849331', 'cg26849382', '\n",
      "1000 input query terms found no hit:\t['cg26883264', 'cg26883282', 'cg26883351', 'cg26883434', 'cg26883472', 'cg26883517', 'cg26883650', '\n",
      "1000 input query terms found no hit:\t['cg26916888', 'cg26916894', 'cg26916916', 'cg26916927', 'cg26916936', 'cg26916966', 'cg26917068', '\n",
      "1000 input query terms found no hit:\t['cg26949731', 'cg26949796', 'cg26949829', 'cg26949883', 'cg26950022', 'cg26950080', 'cg26950102', '\n",
      "1000 input query terms found no hit:\t['cg26983198', 'cg26983228', 'cg26983249', 'cg26983317', 'cg26983326', 'cg26983360', 'cg26983371', '\n",
      "1000 input query terms found no hit:\t['cg27018485', 'cg27018502', 'cg27018554', 'cg27018587', 'cg27018849', 'cg27018904', 'cg27018907', '\n",
      "1000 input query terms found no hit:\t['cg27052428', 'cg27052442', 'cg27052455', 'cg27052498', 'cg27052537', 'cg27052564', 'cg27052575', '\n",
      "1000 input query terms found no hit:\t['cg27086006', 'cg27086014', 'cg27086020', 'cg27086024', 'cg27086028', 'cg27086051', 'cg27086116', '\n",
      "1000 input query terms found no hit:\t['cg27118059', 'cg27118062', 'cg27118072', 'cg27118080', 'cg27118102', 'cg27118178', 'cg27118183', '\n",
      "1000 input query terms found no hit:\t['cg27153146', 'cg27153155', 'cg27153194', 'cg27153316', 'cg27153327', 'cg27153400', 'cg27153492', '\n",
      "1000 input query terms found no hit:\t['cg27187848', 'cg27187881', 'cg27187909', 'cg27188001', 'cg27188026', 'cg27188050', 'cg27188056', '\n",
      "1000 input query terms found no hit:\t['cg27226360', 'cg27226418', 'cg27226424', 'cg27226427', 'cg27226479', 'cg27226511', 'cg27226525', '\n",
      "1000 input query terms found no hit:\t['cg27262488', 'cg27262671', 'cg27262702', 'cg27262717', 'cg27262735', 'cg27262743', 'cg27262745', '\n",
      "1000 input query terms found no hit:\t['cg27297705', 'cg27297722', 'cg27297724', 'cg27297753', 'cg27297761', 'cg27297851', 'cg27297893', '\n",
      "1000 input query terms found no hit:\t['cg27327475', 'cg27327551', 'cg27327580', 'cg27327588', 'cg27327591', 'cg27327598', 'cg27327627', '\n",
      "1000 input query terms found no hit:\t['cg27361847', 'cg27361865', 'cg27361914', 'cg27361933', 'cg27361961', 'cg27361964', 'cg27362010', '\n",
      "1000 input query terms found no hit:\t['cg27394946', 'cg27394962', 'cg27394987', 'cg27395029', 'cg27395066', 'cg27395087', 'cg27395129', '\n",
      "1000 input query terms found no hit:\t['cg27429258', 'cg27429305', 'cg27429313', 'cg27429364', 'cg27429450', 'cg27429457', 'cg27429480', '\n",
      "1000 input query terms found no hit:\t['cg27464670', 'cg27464711', 'cg27464733', 'cg27464755', 'cg27464877', 'cg27465141', 'cg27465151', '\n",
      "1000 input query terms found no hit:\t['cg27500322', 'cg27500384', 'cg27500386', 'cg27500431', 'cg27500473', 'cg27500519', 'cg27500542', '\n",
      "1000 input query terms found no hit:\t['cg27532919', 'cg27532938', 'cg27532955', 'cg27533013', 'cg27533019', 'cg27533080', 'cg27533102', '\n",
      "1000 input query terms found no hit:\t['cg27565609', 'cg27565650', 'cg27565719', 'cg27565794', 'cg27565803', 'cg27565811', 'cg27565852', '\n",
      "1000 input query terms found no hit:\t['cg27600082', 'cg27600084', 'cg27600107', 'cg27600134', 'cg27600136', 'cg27600161', 'cg27600205', '\n",
      "1000 input query terms found no hit:\t['cg27634501', 'cg27634549', 'cg27634629', 'cg27634633', 'cg27634682', 'cg27634695', 'cg27634724', '\n",
      "1000 input query terms found no hit:\t['ch.1.194336624R', 'ch.1.194404001R', 'ch.1.194862R', 'ch.1.196707923R', 'ch.1.196733536R', 'ch.1.1\n",
      "1000 input query terms found no hit:\t['ch.16.51328157R', 'ch.16.53179517R', 'ch.16.5395104F', 'ch.16.54217905R', 'ch.16.55305377R', 'ch.1\n",
      "859 input query terms found no hit:\t['ch.4.58175585R', 'ch.4.58247647F', 'ch.4.58270376F', 'ch.4.586044R', 'ch.4.594573R', 'ch.4.6064613\n"
     ]
    }
   ],
   "source": [
    "if NORMALIZE_GENE:\n",
    "    genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.572778564Z",
     "start_time": "2024-01-10T21:40:20.565101065Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "# The preprocessing runs through, which means is_available should be True\n",
    "is_available = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595385887Z",
     "start_time": "2024-01-10T21:40:20.574630883Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 32 samples.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.595934649Z",
     "start_time": "2024-01-10T21:40:20.580408525Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the feature 'Arrhythmia', the least common label is '1.0' with 32 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Arrhythmia' in this dataset is severely biased.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 32 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.596022151Z",
     "start_time": "2024-01-10T21:40:20.588443384Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.681345468Z",
     "start_time": "2024-01-10T21:40:20.591286171Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:22.271683755Z",
     "start_time": "2023-12-31T03:58:22.246557674Z"
    },
    "id": "-MTPhRGxJV7I"
   },
   "source": [
    "### 3. Do regression & Cross Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.694597045Z",
     "start_time": "2024-01-10T21:40:20.642161512Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the information of usable cohorts\n",
    "best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH)\n",
    "ranked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:20.907514912Z",
     "start_time": "2024-01-10T21:40:20.907178250Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# If both age and gender have available cohorts, select 'age' as the condition.\n",
    "condition = 'Age'\n",
    "filter_column = 'has_' + condition.lower()\n",
    "\n",
    "condition_best_cohort, condition_ranked_df = filter_and_rank_cohorts(JSON_PATH, filter_column)\n",
    "condition_best_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-01-10T21:40:21.120075804Z",
     "start_time": "2024-01-10T21:40:21.109376558Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>is_usable</th>\n",
       "      <th>is_available</th>\n",
       "      <th>is_biased</th>\n",
       "      <th>has_age</th>\n",
       "      <th>has_gender</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [cohort_id, is_usable, is_available, is_biased, has_age, has_gender, sample_size, note]\n",
       "Index: []"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_ranked_df.head()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
