{"cells":[{"cell_type":"markdown","metadata":{"collapsed":false,"id":"8Q7_HywEiaDx"},"source":["# Gold standard curation: Preprocessing and single-step regression"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"pCI2xoVwiaDy"},"source":["In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n","\n","1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n","2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"3lP_4rV0iaDz"},"source":["# 1. Basic setup"]},{"cell_type":"code","execution_count":33,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:07.313098989Z","start_time":"2023-12-31T21:19:07.312718292Z"},"executionInfo":{"elapsed":1480,"status":"ok","timestamp":1704459407389,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"LqmCfkN7iaDz"},"outputs":[],"source":["import os\n","import sys\n","\n","sys.path.append('..')\n","from utils import *\n","# Set your preferred name\n","USER = \"Arthur\"\n","# Set the data and output directories\n","DATA_ROOT = 'C:/Users/arthu/Downloads'\n","OUTPUT_ROOT = 'C:/Users/arthu/OneDrive/Documents/GitHub/output2'\n","TRAIT = \"Autoinflammatory Disorders\"\n","\n","OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n","JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n","if not os.path.exists(OUTPUT_DIR):\n","    os.makedirs(OUTPUT_DIR, exist_ok=True)\n","\n","# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n","NORMALIZE_GENE = True"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"iNY8dzJAiaD0"},"source":["# 2. Data preprocessing and selection"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"7aNtx3jjiaD1"},"source":["## 2.1. The TCGA Xena dataset"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"nn68pLQ1iaD1"},"source":["In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."]},{"cell_type":"code","execution_count":4,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.770095084Z","start_time":"2023-12-31T21:19:09.102451553Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3886,"status":"ok","timestamp":1704459428432,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"hCwvC9PNiaD1","outputId":"9195a4d8-b7ac-4910-d90a-1a3bddba33d9"},"outputs":[{"data":{"text/plain":["['.DS_Store',\n"," 'TCGA_Esophageal_Cancer_(ESCA)',\n"," 'TCGA_Pheochromocytoma_Paraganglioma_(PCPG)',\n"," 'TCGA_Bile_Duct_Cancer_(CHOL)',\n"," 'TCGA_Kidney_Chromophobe_(KICH)',\n"," 'TCGA_Thyroid_Cancer_(THCA)',\n"," 'TCGA_Sarcoma_(SARC)',\n"," 'TCGA_Adrenocortical_Cancer_(ACC)',\n"," 'TCGA_Thymoma_(THYM)',\n"," 'TCGA_Glioblastoma_(GBM)']"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["import os\n","from utils import *\n","\n","dataset = 'TCGA'\n","dataset_dir = os.path.join(DATA_ROOT, dataset)\n","os.listdir(dataset_dir)[:10]"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"sksNMtjOiaD1"},"source":["If no match is found, jump directly to GEO in Part 2.2"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.918243934Z","start_time":"2023-12-31T21:19:09.909021135Z"},"id":"hHU-R89oiaD1"},"outputs":[],"source":["trait_subdir = \"TCGA_Breast_Cancer_(BRCA)\"\n","cohort = 'Xena'\n","# All the cancer traits in Xena are binary\n","trait_type = 'binary'\n","# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n","is_available = True\n","\n","clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_BRCA_clinicalMatrix')\n","genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_HiSeqV2_PANCAN.gz')"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.613198409Z","start_time":"2023-12-31T21:19:10.409688003Z"},"id":"MudwB-_iz7sc"},"outputs":[],"source":["import pandas as pd\n","\n","clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n","genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n","age_col = gender_col = None"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.617006167Z","start_time":"2023-12-31T21:19:13.614080653Z"},"id":"8XwEOygHiaD2"},"outputs":[],"source":["_, clinical_data_cols = check_rows_and_columns(clinical_data)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661538569Z","start_time":"2023-12-31T21:19:13.615946213Z"},"id":"TsoPD93diaD2","outputId":"b7b9aa38-4f63-4eea-8648-f208c80a4a60"},"outputs":[{"data":{"text/plain":["['AJCC_Stage_nature2012',\n"," 'Age_at_Initial_Pathologic_Diagnosis_nature2012',\n"," 'CN_Clusters_nature2012',\n"," 'Converted_Stage_nature2012',\n"," 'Days_to_Date_of_Last_Contact_nature2012',\n"," 'Days_to_date_of_Death_nature2012',\n"," 'ER_Status_nature2012',\n"," 'Gender_nature2012',\n"," 'HER2_Final_Status_nature2012',\n"," 'Integrated_Clusters_no_exp__nature2012']"]},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data_cols[:10]"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-28T20:20:11.841042813Z","start_time":"2023-12-28T20:20:11.834458206Z"},"collapsed":false,"id":"Ejzdxb0diaD2"},"source":["Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n","Reference prompt:"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661858700Z","start_time":"2023-12-31T21:19:13.656749487Z"},"id":"X11W7KfziaD2","outputId":"ef120e77-44dd-443b-f86b-e464d962574c"},"outputs":[{"data":{"text/plain":["\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['AJCC_Stage_nature2012', 'Age_at_Initial_Pathologic_Diagnosis_nature2012', 'CN_Clusters_nature2012', 'Converted_Stage_nature2012', 'Days_to_Date_of_Last_Contact_nature2012', 'Days_to_date_of_Death_nature2012', 'ER_Status_nature2012', 'Gender_nature2012', 'HER2_Final_Status_nature2012', 'Integrated_Clusters_no_exp__nature2012', 'Integrated_Clusters_unsup_exp__nature2012', 'Integrated_Clusters_with_PAM50__nature2012', 'Metastasis_Coded_nature2012', 'Metastasis_nature2012', 'Node_Coded_nature2012', 'Node_nature2012', 'OS_Time_nature2012', 'OS_event_nature2012', 'PAM50Call_RNAseq', 'PAM50_mRNA_nature2012', 'PR_Status_nature2012', 'RPPA_Clusters_nature2012', 'SigClust_Intrinsic_mRNA_nature2012', 'SigClust_Unsupervised_mRNA_nature2012', 'Survival_Data_Form_nature2012', 'Tumor_T1_Coded_nature2012', 'Tumor_nature2012', 'Vital_Status_nature2012', '_INTEGRATION', '_PANCAN_CNA_PANCAN_K8', '_PANCAN_Cluster_Cluster_PANCAN', '_PANCAN_DNAMethyl_BRCA', '_PANCAN_DNAMethyl_PANCAN', '_PANCAN_RPPA_PANCAN_K8', '_PANCAN_UNC_RNAseq_PANCAN_K16', '_PANCAN_miRNA_PANCAN', '_PANCAN_mirna_BRCA', '_PANCAN_mutation_PANCAN', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_surgery_locoregional_procedure', 'additional_surgery_metastatic_procedure', 'age_at_initial_pathologic_diagnosis', 'anatomic_neoplasm_subdivision', 'axillary_lymph_node_stage_method_type', 'axillary_lymph_node_stage_other_method_descriptive_text', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'breast_cancer_surgery_margin_status', 'breast_carcinoma_estrogen_receptor_status', 'breast_carcinoma_immunohistochemistry_er_pos_finding_scale', 'breast_carcinoma_immunohistochemistry_pos_cell_score', 'breast_carcinoma_immunohistochemistry_prgstrn_rcptr_ps_fndng_scl', 'breast_carcinoma_primary_surgical_procedure_name', 'breast_carcinoma_progesterone_receptor_status', 'breast_carcinoma_surgical_procedure_name', 'breast_neoplasm_other_surgical_procedure_descriptive_text', 'cytokeratin_immunohistochemistry_staining_method_mcrmtstss_ndctr', 'days_to_additional_surgery_locoregional_procedure', 'days_to_additional_surgery_metastatic_procedure', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_last_known_alive', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'disease_code', 'distant_metastasis_present_ind2', 'er_detection_method_text', 'er_level_cell_percentage_category', 'fluorescence_in_st_hybrdztn_dgnstc_prcdr_chrmsm_17_sgnl_rslt_rng', 'followup_case_report_form_submission_reason', 'form_completion_date', 'gender', 'her2_and_centromere_17_positive_finding_other_measuremnt_scl_txt', 'her2_erbb_method_calculation_method_text', 'her2_erbb_pos_finding_cell_percent_category', 'her2_erbb_pos_finding_fluorescence_n_st_hybrdztn_clcltn_mthd_txt', 'her2_immunohistochemistry_level_result', 'her2_neu_and_centromere_17_copy_number_analysis_npt_ttl_nmbr_cnt', 'her2_neu_breast_carcinoma_copy_analysis_input_total_number', 'her2_neu_chromosone_17_signal_ratio_value', 'her2_neu_metastatic_breast_carcinoma_copy_analysis_inpt_ttl_nmbr', 'histological_type', 'histological_type_other', 'history_of_neoadjuvant_treatment', 'hr2_n_nd_cntrmr_17_cpy_nmbr_mtsttc_brst_crcnm_nlyss_npt_ttl_nmbr', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'immunohistochemistry_positive_cell_score', 'informed_consent_verified', 'init_pathology_dx_method_other', 'initial_pathologic_diagnosis_method', 'initial_weight', 'is_ffpe', 'lab_proc_her2_neu_immunohistochemistry_receptor_status', 'lab_procedure_her2_neu_in_situ_hybrid_outcome_type', 'lost_follow_up', 'lymph_node_examined_count', 'margin_status', 'menopause_status', 'metastatic_breast_carcinm_ps_fndng_prgstrn_rcptr_thr_msr_scl_txt', 'metastatic_breast_carcinom_lb_prc_hr2_n_mmnhstchmstry_rcptr_stts', 'metastatic_breast_carcinoma_erbb2_immunohistochemistry_levl_rslt', 'metastatic_breast_carcinoma_estrogen_receptor_detection_mthd_txt', 'metastatic_breast_carcinoma_estrogen_receptor_status', 'metastatic_breast_carcinoma_estrogen_receptr_lvl_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_her2_erbb_method_calculatin_mthd_txt', 'metastatic_breast_carcinoma_her2_erbb_pos_findng_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_her2_neu_chromosone_17_signal_rat_vl', 'metastatic_breast_carcinoma_immunhstchmstry_r_pstv_fndng_scl_typ', 'metastatic_breast_carcinoma_immunohistochemistry_er_pos_cell_scr', 'metastatic_breast_carcinoma_immunohistochemistry_pr_pos_cell_scr', 'metastatic_breast_carcinoma_lab_proc_hr2_n_n_st_hybrdztn_tcm_typ', 'metastatic_breast_carcinoma_pos_finding_hr2_rbb2_thr_msr_scl_txt', 'metastatic_breast_carcinoma_progestern_rcptr_lvl_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_progesterone_receptor_dtctn_mthd_txt', 'metastatic_breast_carcinoma_progesterone_receptor_status', 'metastatic_site_at_diagnosis', 'metastatic_site_at_diagnosis_other', 'methylation_Clusters_nature2012', 'miRNA_Clusters_nature2012', 'mtsttc_brst_crcnm_flrscnc_n_st_hybrdztn_dgnstc_prc_cntrmr_17_sgn', 'mtsttc_brst_crcnm_hr2_rbb_ps_fndng_flrscnc_n_st_hybrdztn_clcltn', 'mtsttc_brst_crcnm_mmnhstchmstry_prgstrn_rcptr_pstv_fndng_scl_typ', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'number_of_lymphnodes_positive_by_he', 'number_of_lymphnodes_positive_by_ihc', 'oct_embedded', 'other_dx', 'pathologic_M', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'person_neoplasm_cancer_status', 'pgr_detection_method_text', 'pos_finding_her2_erbb2_other_measurement_scale_text', 'pos_finding_metastatic_brst_crcnm_strgn_rcptr_thr_msrmnt_scl_txt', 'pos_finding_progesterone_receptor_other_measurement_scale_text', 'positive_finding_estrogen_receptor_other_measurement_scale_text', 'postoperative_rx_tx', 'primary_lymph_node_presentation_assessment', 'progesterone_receptor_level_cell_percent_category', 'project_code', 'radiation_therapy', 'sample_type', 'sample_type_id', 'surgical_procedure_purpose_other_text', 'system_version', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'vial_number', 'vital_status', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_BRCA_RPPA_RBN', '_GENOMIC_ID_TCGA_BRCA_mutation', '_GENOMIC_ID_TCGA_BRCA_PDMRNAseq', '_GENOMIC_ID_TCGA_BRCA_hMethyl450', '_GENOMIC_ID_TCGA_BRCA_RPPA', '_GENOMIC_ID_TCGA_BRCA_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_BRCA_mutation_curated_wustl_gene', '_GENOMIC_ID_TCGA_BRCA_hMethyl27', '_GENOMIC_ID_TCGA_BRCA_PDMarrayCNV', '_GENOMIC_ID_TCGA_BRCA_miRNA_HiSeq', '_GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene', '_GENOMIC_ID_TCGA_BRCA_miRNA_GA', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile', '_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene', '_GENOMIC_ID_TCGA_BRCA_gistic2thd', '_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_BRCA_G4502A_07_3', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2', '_GENOMIC_ID_TCGA_BRCA_gistic2', '_GENOMIC_ID_TCGA_BRCA_PDMarray']\\n\""]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["f'''\n","Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n","candidate_age_cols = [col_name1, col_name2, ...]\n","candidate_gender_cols = [col_name1, col_name2, ...]\n","If no columns match a criterion, please provide an empty list.\n","\n","Column names:\n","{clinical_data_cols}\n","'''"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661938833Z","start_time":"2023-12-31T21:19:13.656901848Z"},"id":"SicVNjZ4iaD2"},"outputs":[],"source":["candidate_age_cols = ['Age_at_Initial_Pathologic_Diagnosis_nature2012', 'age_at_initial_pathologic_diagnosis',\n","                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n","candidate_gender_cols = ['Gender_nature2012', 'gender']"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:57:44.207565572Z","start_time":"2023-12-31T03:57:44.202177544Z"},"collapsed":false,"id":"APRmqURGiaD2"},"source":["Choose a single column from the candidate columns that record age and gender information respectively.\n","If no column meets the requirement, keep 'age_col' or 'gender_col' to None"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.662078139Z","start_time":"2023-12-31T21:19:13.657006017Z"},"id":"K6uQ_zrZiaD3","outputId":"1602b911-ac43-4f36-8ff6-5366b7bc573a"},"outputs":[{"data":{"text/plain":["{'Age_at_Initial_Pathologic_Diagnosis_nature2012': [nan, nan, nan, nan, nan],\n"," 'age_at_initial_pathologic_diagnosis': [55.0, 50.0, 62.0, 52.0, 50.0],\n"," 'days_to_birth': [-20211.0, -18538.0, -22848.0, -19074.0, -18371.0],\n"," 'year_of_initial_pathologic_diagnosis': [2004.0,\n","  2003.0,\n","  2011.0,\n","  2011.0,\n","  2013.0]}"]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_age_cols])"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.777159137Z","start_time":"2023-12-31T21:19:13.756210135Z"},"id":"ks4TtmqNiaD3"},"outputs":[],"source":["age_col = 'age_at_initial_pathologic_diagnosis'"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.313425023Z","start_time":"2023-12-31T21:19:14.305221062Z"},"id":"piIf5weZiaD3","outputId":"08a17b50-20d8-41bd-e1c7-fd8020193ceb"},"outputs":[{"data":{"text/plain":["{'Gender_nature2012': [nan, nan, nan, nan, nan],\n"," 'gender': ['FEMALE', 'FEMALE', 'FEMALE', 'FEMALE', 'FEMALE']}"]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_gender_cols])"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.852502962Z","start_time":"2023-12-31T21:19:14.845153121Z"},"id":"tPNM-3cdiaD3"},"outputs":[],"source":["gender_col = 'gender'"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.246065025Z","start_time":"2023-12-31T21:19:15.240279162Z"},"id":"Ed3TUm8_iaD3"},"outputs":[],"source":["selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col, gender_col)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.593478652Z","start_time":"2023-12-31T21:19:15.583841656Z"},"id":"4BDE6fnKiaD3"},"outputs":[],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.074598194Z","start_time":"2023-12-31T21:19:15.973600157Z"},"id":"Kao1VPxfiaD3","outputId":"a072e535-52f6-4ebf-8ba6-c2f59accfea7"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Breast Cancer</th>\n","      <th>Age</th>\n","      <th>Gender</th>\n","      <th>ARHGEF10L</th>\n","      <th>HIF3A</th>\n","      <th>RNF17</th>\n","      <th>RNF10</th>\n","      <th>RNF11</th>\n","      <th>RNF13</th>\n","      <th>GTF2IP1</th>\n","      <th>...</th>\n","      <th>TULP2</th>\n","      <th>NPY5R</th>\n","      <th>GNGT2</th>\n","      <th>GNGT1</th>\n","      <th>TULP3</th>\n","      <th>PTRF</th>\n","      <th>BCL6B</th>\n","      <th>GSTK1</th>\n","      <th>SELP</th>\n","      <th>SELS</th>\n","    </tr>\n","    <tr>\n","      <th>sampleID</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>TCGA-3C-AAAU-01</th>\n","      <td>1</td>\n","      <td>55.0</td>\n","      <td>0.0</td>\n","      <td>0.607308</td>\n","      <td>-3.194126</td>\n","      <td>-0.531035</td>\n","      <td>-0.145872</td>\n","      <td>0.237422</td>\n","      <td>-0.29921</td>\n","      <td>-0.142694</td>\n","      <td>...</td>\n","      <td>-0.748878</td>\n","      <td>0.565583</td>\n","      <td>-0.767233</td>\n","      <td>-1.28139</td>\n","      <td>-0.271377</td>\n","      <td>-0.492286</td>\n","      <td>0.360373</td>\n","      <td>1.067905</td>\n","      <td>0.076267</td>\n","      <td>-0.392212</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-3C-AALI-01</th>\n","      <td>1</td>\n","      <td>50.0</td>\n","      <td>0.0</td>\n","      <td>-0.641192</td>\n","      <td>-4.928226</td>\n","      <td>0.095465</td>\n","      <td>0.098128</td>\n","      <td>-0.541978</td>\n","      <td>-0.32291</td>\n","      <td>-0.044694</td>\n","      <td>...</td>\n","      <td>0.312922</td>\n","      <td>0.079683</td>\n","      <td>0.475267</td>\n","      <td>1.99661</td>\n","      <td>0.161423</td>\n","      <td>0.273714</td>\n","      <td>0.644673</td>\n","      <td>-0.428695</td>\n","      <td>0.068667</td>\n","      <td>-0.043812</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-3C-AALJ-01</th>\n","      <td>1</td>\n","      <td>62.0</td>\n","      <td>0.0</td>\n","      <td>1.082808</td>\n","      <td>-4.623726</td>\n","      <td>-0.531035</td>\n","      <td>0.484028</td>\n","      <td>-0.183678</td>\n","      <td>-0.91901</td>\n","      <td>0.261106</td>\n","      <td>...</td>\n","      <td>-0.748878</td>\n","      <td>-0.656117</td>\n","      <td>-0.216733</td>\n","      <td>0.21081</td>\n","      <td>0.101023</td>\n","      <td>0.660514</td>\n","      <td>1.295073</td>\n","      <td>0.915105</td>\n","      <td>0.168567</td>\n","      <td>0.047788</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-3C-AALK-01</th>\n","      <td>1</td>\n","      <td>52.0</td>\n","      <td>0.0</td>\n","      <td>0.121608</td>\n","      <td>-2.881526</td>\n","      <td>-0.531035</td>\n","      <td>0.179128</td>\n","      <td>0.039222</td>\n","      <td>-0.45491</td>\n","      <td>0.180306</td>\n","      <td>...</td>\n","      <td>0.415422</td>\n","      <td>-0.178417</td>\n","      <td>-0.211233</td>\n","      <td>-1.28139</td>\n","      <td>0.108023</td>\n","      <td>1.132814</td>\n","      <td>0.496773</td>\n","      <td>0.240105</td>\n","      <td>3.099767</td>\n","      <td>0.112888</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-4H-AAAK-01</th>\n","      <td>1</td>\n","      <td>50.0</td>\n","      <td>0.0</td>\n","      <td>0.420208</td>\n","      <td>-3.282726</td>\n","      <td>-0.531035</td>\n","      <td>-0.020972</td>\n","      <td>-0.117978</td>\n","      <td>-0.55781</td>\n","      <td>-0.173794</td>\n","      <td>...</td>\n","      <td>0.685222</td>\n","      <td>-0.698717</td>\n","      <td>-0.484233</td>\n","      <td>-1.28139</td>\n","      <td>-0.223577</td>\n","      <td>1.222714</td>\n","      <td>0.225573</td>\n","      <td>-0.301995</td>\n","      <td>0.200067</td>\n","      <td>-0.080212</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 20533 columns</p>\n","</div>"],"text/plain":["                 Breast Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17  \\\n","sampleID                                                                      \n","TCGA-3C-AAAU-01              1  55.0     0.0   0.607308 -3.194126 -0.531035   \n","TCGA-3C-AALI-01              1  50.0     0.0  -0.641192 -4.928226  0.095465   \n","TCGA-3C-AALJ-01              1  62.0     0.0   1.082808 -4.623726 -0.531035   \n","TCGA-3C-AALK-01              1  52.0     0.0   0.121608 -2.881526 -0.531035   \n","TCGA-4H-AAAK-01              1  50.0     0.0   0.420208 -3.282726 -0.531035   \n","\n","                    RNF10     RNF11    RNF13   GTF2IP1  ...     TULP2  \\\n","sampleID                                                ...             \n","TCGA-3C-AAAU-01 -0.145872  0.237422 -0.29921 -0.142694  ... -0.748878   \n","TCGA-3C-AALI-01  0.098128 -0.541978 -0.32291 -0.044694  ...  0.312922   \n","TCGA-3C-AALJ-01  0.484028 -0.183678 -0.91901  0.261106  ... -0.748878   \n","TCGA-3C-AALK-01  0.179128  0.039222 -0.45491  0.180306  ...  0.415422   \n","TCGA-4H-AAAK-01 -0.020972 -0.117978 -0.55781 -0.173794  ...  0.685222   \n","\n","                    NPY5R     GNGT2    GNGT1     TULP3      PTRF     BCL6B  \\\n","sampleID                                                                     \n","TCGA-3C-AAAU-01  0.565583 -0.767233 -1.28139 -0.271377 -0.492286  0.360373   \n","TCGA-3C-AALI-01  0.079683  0.475267  1.99661  0.161423  0.273714  0.644673   \n","TCGA-3C-AALJ-01 -0.656117 -0.216733  0.21081  0.101023  0.660514  1.295073   \n","TCGA-3C-AALK-01 -0.178417 -0.211233 -1.28139  0.108023  1.132814  0.496773   \n","TCGA-4H-AAAK-01 -0.698717 -0.484233 -1.28139 -0.223577  1.222714  0.225573   \n","\n","                    GSTK1      SELP      SELS  \n","sampleID                                       \n","TCGA-3C-AAAU-01  1.067905  0.076267 -0.392212  \n","TCGA-3C-AALI-01 -0.428695  0.068667 -0.043812  \n","TCGA-3C-AALJ-01  0.915105  0.168567  0.047788  \n","TCGA-3C-AALK-01  0.240105  3.099767  0.112888  \n","TCGA-4H-AAAK-01 -0.301995  0.200067 -0.080212  \n","\n","[5 rows x 20533 columns]"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n","merged_data.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.341121723Z","start_time":"2023-12-31T21:19:16.330544222Z"},"id":"x7E8M0H-iaD3","outputId":"c14ee046-590a-4ea8-fcbb-1fcadca55632"},"outputs":[{"name":"stdout","output_type":"stream","text":["The merged dataset contains 1215 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:17.222505900Z","start_time":"2023-12-31T21:19:17.218781412Z"},"id":"YvBbzAe6iaD3","outputId":"15cf8038-a874-482d-af3d-d1204569e905"},"outputs":[{"name":"stdout","output_type":"stream","text":["For the feature 'Breast Cancer', the least common label is '0' with 113 occurrences. This represents 9.30% of the dataset.\n","The distribution of the feature 'Breast Cancer' in this dataset is fine.\n","\n","Quartiles for 'Age':\n","  25%: 48.0\n","  50% (Median): 58.0\n","  75%: 67.0\n","Min: 26.0\n","Max: 90.0\n","The distribution of the feature 'Age' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '1.0' with 13 occurrences. This represents 1.07% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"data":{"text/plain":["False"]},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.401536463Z","start_time":"2023-12-31T21:19:17.980683628Z"},"id":"Qzem-9-CiaD4"},"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.405288661Z","start_time":"2023-12-31T21:19:35.402799553Z"},"id":"6LuV7GjKiaD4"},"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-28T21:28:38.028916303Z","start_time":"2023-12-28T21:28:38.016245426Z"},"collapsed":false,"id":"Kk0EYJ4yiaD4"},"source":["## 2.2. The GEO dataset"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:25:23.253882615Z","start_time":"2023-12-31T03:25:23.244062710Z"},"collapsed":false,"id":"-RCIZQ6jiaD4"},"source":["In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."]},{"cell_type":"code","execution_count":2,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.411177575Z","start_time":"2023-12-31T21:19:35.405710636Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2642,"status":"ok","timestamp":1704459441002,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"zLCP2SDmiaD4","outputId":"fe593ad4-5134-410b-f0b2-e8e32b11e452"},"outputs":[{"data":{"text/plain":["['GSE103500', 'GSE43553', 'GSE80060']"]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":["dataset = 'GEO'\n","trait_subdir = \"Autoinflammatory-Disorders\"\n","\n","trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n","os.listdir(trait_path)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:39:42.634870142Z","start_time":"2023-12-31T03:39:42.534093295Z"},"collapsed":false,"id":"VHR6J86PiaD4"},"source":["Repeat the below steps for all the accession numbers"]},{"cell_type":"code","execution_count":135,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.413836016Z","start_time":"2023-12-31T21:19:35.410443976Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704459441002,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"-NttBC9AiaD4","outputId":"83884606-97d8-428c-9e4f-0c60104f30bf"},"outputs":[{"data":{"text/plain":["('C:/Users/arthu/Downloads\\\\GEO\\\\Autoinflammatory-Disorders\\\\GSE80060\\\\GSE80060_family.soft.gz',\n"," 'C:/Users/arthu/Downloads\\\\GEO\\\\Autoinflammatory-Disorders\\\\GSE80060\\\\GSE80060_series_matrix.txt.gz')"]},"execution_count":135,"metadata":{},"output_type":"execute_result"}],"source":["cohort = accession_num = \"GSE80060\"\n","cohort_dir = os.path.join(trait_path, accession_num)\n","soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n","soft_file, matrix_file"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"OXII6n-eiaD4"},"source":["### Inital filtering and clinical data preprocessing"]},{"cell_type":"code","execution_count":136,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.492119842Z","start_time":"2023-12-31T21:19:35.413383102Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1920,"status":"ok","timestamp":1704459444880,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"zLBtCC4eiaD4","outputId":"e9955b06-f42c-4711-cd72-6b2294ed5ca4"},"outputs":[{"name":"stdout","output_type":"stream","text":["!Series_title\t\"Gene expression data of whole blood of systemic juvenile idiopathic arthritis (SJIA) patients treated with canakinumab or placebo and age matched healthy controls\"\n","!Series_summary\t\"Canakinumab is a human anti-interleukin-1 beta (IL-1 beta) monoclonal antibody neutralizing IL-1 beta. Systemic juvenile idiopathic arthritis (SJIA) is a rare, multigenic, autoinflammatory disease of unknown etiology characterized by chronic arthritis; intermittent high-spiking fever, rash, and elevated levels of acute-phase reactants. Blood samples of SJIA patients were obtained from two phase 3 clinical trials conducted by the members of the Pediatric Rheumatology International Trials Organization (PRINTO) and the Pediatric Rheumatology Collaborative Study Group (PRCSG) (Clinicaltrials.gov: NCT00886769 and NCT00889863). For patients, baseline and day 3 samples were analyzed for either placebo or canakinumab (Ilaris) treatment.\"\n","!Series_summary\t\"Clinical response was assessed at day 15 using adapted JIA American College of Rheumatology (ACR) response criteria.\"\n","!Series_overall_design\t\"Overall, 206 samples were used in this study including 22 samples from healthy controls, 33 samples of placebo treated patients and 151 samples of canakinumab treated patients.\"\n"]}],"source":["background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n","clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n","\n","background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n","print(background_info)"]},{"cell_type":"code","execution_count":137,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.507840422Z","start_time":"2023-12-31T21:19:35.451789107Z"},"colab":{"base_uri":"https://localhost:8080/","height":360},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1704459446314,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"YwC2YMZ_iaD5","outputId":"790b0033-d324-44a9-a49a-01eefe1aeb45"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>!Sample_geo_accession</th>\n","      <th>GSM2111993</th>\n","      <th>GSM2111994</th>\n","      <th>GSM2111995</th>\n","      <th>GSM2111996</th>\n","      <th>GSM2111997</th>\n","      <th>GSM2111998</th>\n","      <th>GSM2111999</th>\n","      <th>GSM2112000</th>\n","      <th>GSM2112001</th>\n","      <th>...</th>\n","      <th>GSM2112189</th>\n","      <th>GSM2112190</th>\n","      <th>GSM2112191</th>\n","      <th>GSM2112192</th>\n","      <th>GSM2112193</th>\n","      <th>GSM2112194</th>\n","      <th>GSM2112195</th>\n","      <th>GSM2112196</th>\n","      <th>GSM2112197</th>\n","      <th>GSM2112198</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>...</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","      <td>tissue: Whole blood</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>...</td>\n","      <td>disease status: Healthy</td>\n","      <td>disease status: Healthy</td>\n","      <td>disease status: Healthy</td>\n","      <td>disease status: Healthy</td>\n","      <td>disease status: Healthy</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","      <td>disease status: SJIA</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>subject id: SJIA_2_2513</td>\n","      <td>subject id: SJIA_2_2513</td>\n","      <td>subject id: SJIA_2_313</td>\n","      <td>subject id: SJIA_2_413</td>\n","      <td>subject id: SJIA_2_413</td>\n","      <td>subject id: SJIA_2_712</td>\n","      <td>subject id: SJIA_2_812</td>\n","      <td>subject id: SJIA_2_912</td>\n","      <td>subject id: SJIA_2_1013</td>\n","      <td>...</td>\n","      <td>subject id: Healthy_subject_19</td>\n","      <td>subject id: Healthy_subject_20</td>\n","      <td>subject id: Healthy_subject_21</td>\n","      <td>subject id: Healthy_subject_22</td>\n","      <td>subject id: Healthy_subject_23</td>\n","      <td>subject id: SJIA_1_3528</td>\n","      <td>subject id: SJIA_1_3630</td>\n","      <td>subject id: SJIA_1_3630</td>\n","      <td>subject id: SJIA_1_3730</td>\n","      <td>subject id: SJIA_1_3730</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day3</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day3</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day3</td>\n","      <td>visit: Day1_BL</td>\n","      <td>...</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day3</td>\n","      <td>visit: Day1_BL</td>\n","      <td>visit: Day3</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Placebo</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Placebo</td>\n","      <td>treatment: Placebo</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>...</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","      <td>treatment: Canakinumab</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 207 columns</p>\n","</div>"],"text/plain":["         !Sample_geo_accession               GSM2111993  \\\n","0  !Sample_characteristics_ch1      tissue: Whole blood   \n","1  !Sample_characteristics_ch1     disease status: SJIA   \n","2  !Sample_characteristics_ch1  subject id: SJIA_2_2513   \n","3  !Sample_characteristics_ch1           visit: Day1_BL   \n","4  !Sample_characteristics_ch1   treatment: Canakinumab   \n","\n","                GSM2111994              GSM2111995              GSM2111996  \\\n","0      tissue: Whole blood     tissue: Whole blood     tissue: Whole blood   \n","1     disease status: SJIA    disease status: SJIA    disease status: SJIA   \n","2  subject id: SJIA_2_2513  subject id: SJIA_2_313  subject id: SJIA_2_413   \n","3              visit: Day3          visit: Day1_BL          visit: Day1_BL   \n","4   treatment: Canakinumab      treatment: Placebo  treatment: Canakinumab   \n","\n","               GSM2111997              GSM2111998              GSM2111999  \\\n","0     tissue: Whole blood     tissue: Whole blood     tissue: Whole blood   \n","1    disease status: SJIA    disease status: SJIA    disease status: SJIA   \n","2  subject id: SJIA_2_413  subject id: SJIA_2_712  subject id: SJIA_2_812   \n","3             visit: Day3          visit: Day1_BL          visit: Day1_BL   \n","4  treatment: Canakinumab      treatment: Placebo      treatment: Placebo   \n","\n","               GSM2112000               GSM2112001  ...  \\\n","0     tissue: Whole blood      tissue: Whole blood  ...   \n","1    disease status: SJIA     disease status: SJIA  ...   \n","2  subject id: SJIA_2_912  subject id: SJIA_2_1013  ...   \n","3             visit: Day3           visit: Day1_BL  ...   \n","4  treatment: Canakinumab   treatment: Canakinumab  ...   \n","\n","                       GSM2112189                      GSM2112190  \\\n","0             tissue: Whole blood             tissue: Whole blood   \n","1         disease status: Healthy         disease status: Healthy   \n","2  subject id: Healthy_subject_19  subject id: Healthy_subject_20   \n","3                  visit: Day1_BL                  visit: Day1_BL   \n","4                 treatment: none                 treatment: none   \n","\n","                       GSM2112191                      GSM2112192  \\\n","0             tissue: Whole blood             tissue: Whole blood   \n","1         disease status: Healthy         disease status: Healthy   \n","2  subject id: Healthy_subject_21  subject id: Healthy_subject_22   \n","3                  visit: Day1_BL                  visit: Day1_BL   \n","4                 treatment: none                 treatment: none   \n","\n","                       GSM2112193               GSM2112194  \\\n","0             tissue: Whole blood      tissue: Whole blood   \n","1         disease status: Healthy     disease status: SJIA   \n","2  subject id: Healthy_subject_23  subject id: SJIA_1_3528   \n","3                  visit: Day1_BL           visit: Day1_BL   \n","4                 treatment: none   treatment: Canakinumab   \n","\n","                GSM2112195               GSM2112196               GSM2112197  \\\n","0      tissue: Whole blood      tissue: Whole blood      tissue: Whole blood   \n","1     disease status: SJIA     disease status: SJIA     disease status: SJIA   \n","2  subject id: SJIA_1_3630  subject id: SJIA_1_3630  subject id: SJIA_1_3730   \n","3           visit: Day1_BL              visit: Day3           visit: Day1_BL   \n","4   treatment: Canakinumab   treatment: Canakinumab   treatment: Canakinumab   \n","\n","                GSM2112198  \n","0      tissue: Whole blood  \n","1     disease status: SJIA  \n","2  subject id: SJIA_1_3730  \n","3              visit: Day3  \n","4   treatment: Canakinumab  \n","\n","[5 rows x 207 columns]"]},"execution_count":137,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data.head()"]},{"cell_type":"code","execution_count":138,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508015426Z","start_time":"2023-12-31T21:19:35.492746637Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704459448203,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"qUEEMODXiaD5","outputId":"530794df-97b8-40b0-fa0a-f94bf6850ee9"},"outputs":[{"data":{"text/plain":["{0: ['tissue: Whole blood'],\n"," 1: ['disease status: SJIA', 'disease status: Healthy'],\n"," 2: ['subject id: SJIA_2_2513',\n","  'subject id: SJIA_2_313',\n","  'subject id: SJIA_2_413',\n","  'subject id: SJIA_2_712',\n","  'subject id: SJIA_2_812',\n","  'subject id: SJIA_2_912',\n","  'subject id: SJIA_2_1013',\n","  'subject id: SJIA_2_1112',\n","  'subject id: SJIA_2_2912',\n","  'subject id: SJIA_2_3012',\n","  'subject id: SJIA_2_1413',\n","  'subject id: SJIA_2_1411',\n","  'subject id: SJIA_2_168',\n","  'subject id: SJIA_2_167',\n","  'subject id: SJIA_2_1713',\n","  'subject id: SJIA_2_1811',\n","  'subject id: SJIA_2_185',\n","  'subject id: SJIA_2_1912',\n","  'subject id: SJIA_2_2213',\n","  'subject id: SJIA_2_2313',\n","  'subject id: SJIA_2_2312',\n","  'subject id: SJIA_2_113',\n","  'subject id: SJIA_2_2613',\n","  'subject id: SJIA_2_212',\n","  'subject id: SJIA_2_310',\n","  'subject id: SJIA_2_36',\n","  'subject id: SJIA_2_512',\n","  'subject id: SJIA_2_511',\n","  'subject id: SJIA_2_613',\n","  'subject id: SJIA_2_612'],\n"," 3: ['visit: Day1_BL', 'visit: Day3'],\n"," 4: ['treatment: Canakinumab', 'treatment: Placebo', 'treatment: none'],\n"," 5: ['acr response at day 15: 100',\n","  'acr response at day 15: NA',\n","  'acr response at day 15: 30',\n","  'acr response at day 15: 70',\n","  'acr response at day 15: 90',\n","  'acr response at day 15: 0',\n","  'acr response at day 15: 50']}"]},"execution_count":138,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data_unique = get_unique_values_by_row(clinical_data)\n","clinical_data_unique"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978204446Z","start_time":"2023-12-31T03:58:04.922270095Z"},"collapsed":false,"id":"zueh_J2PiaD5"},"source":["Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n","Reference prompt:"]},{"cell_type":"code","execution_count":139,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508232400Z","start_time":"2023-12-31T21:19:35.492903487Z"},"colab":{"base_uri":"https://localhost:8080/","height":178},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1704459451849,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"avSkGIZQiaD_","outputId":"3d9cf239-2a8d-41c1-c111-78cd92d031e4"},"outputs":[{"data":{"text/plain":["'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Autoinflammatory Disorders\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Autoinflammatory Disorders\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Gene expression data of whole blood of systemic juvenile idiopathic arthritis (SJIA) patients treated with canakinumab or placebo and age matched healthy controls\"\\n!Series_summary\\t\"Canakinumab is a human anti-interleukin-1 beta (IL-1 beta) monoclonal antibody neutralizing IL-1 beta. Systemic juvenile idiopathic arthritis (SJIA) is a rare, multigenic, autoinflammatory disease of unknown etiology characterized by chronic arthritis; intermittent high-spiking fever, rash, and elevated levels of acute-phase reactants. Blood samples of SJIA patients were obtained from two phase 3 clinical trials conducted by the members of the Pediatric Rheumatology International Trials Organization (PRINTO) and the Pediatric Rheumatology Collaborative Study Group (PRCSG) (Clinicaltrials.gov: NCT00886769 and NCT00889863). For patients, baseline and day 3 samples were analyzed for either placebo or canakinumab (Ilaris) treatment.\"\\n!Series_summary\\t\"Clinical response was assessed at day 15 using adapted JIA American College of Rheumatology (ACR) response criteria.\"\\n!Series_overall_design\\t\"Overall, 206 samples were used in this study including 22 samples from healthy controls, 33 samples of placebo treated patients and 151 samples of canakinumab treated patients.\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'tissue: Whole blood\\'], 1: [\\'disease status: SJIA\\', \\'disease status: Healthy\\'], 2: [\\'subject id: SJIA_2_2513\\', \\'subject id: SJIA_2_313\\', \\'subject id: SJIA_2_413\\', \\'subject id: SJIA_2_712\\', \\'subject id: SJIA_2_812\\', \\'subject id: SJIA_2_912\\', \\'subject id: SJIA_2_1013\\', \\'subject id: SJIA_2_1112\\', \\'subject id: SJIA_2_2912\\', \\'subject id: SJIA_2_3012\\', \\'subject id: SJIA_2_1413\\', \\'subject id: SJIA_2_1411\\', \\'subject id: SJIA_2_168\\', \\'subject id: SJIA_2_167\\', \\'subject id: SJIA_2_1713\\', \\'subject id: SJIA_2_1811\\', \\'subject id: SJIA_2_185\\', \\'subject id: SJIA_2_1912\\', \\'subject id: SJIA_2_2213\\', \\'subject id: SJIA_2_2313\\', \\'subject id: SJIA_2_2312\\', \\'subject id: SJIA_2_113\\', \\'subject id: SJIA_2_2613\\', \\'subject id: SJIA_2_212\\', \\'subject id: SJIA_2_310\\', \\'subject id: SJIA_2_36\\', \\'subject id: SJIA_2_512\\', \\'subject id: SJIA_2_511\\', \\'subject id: SJIA_2_613\\', \\'subject id: SJIA_2_612\\'], 3: [\\'visit: Day1_BL\\', \\'visit: Day3\\'], 4: [\\'treatment: Canakinumab\\', \\'treatment: Placebo\\', \\'treatment: none\\'], 5: [\\'acr response at day 15: 100\\', \\'acr response at day 15: NA\\', \\'acr response at day 15: 30\\', \\'acr response at day 15: 70\\', \\'acr response at day 15: 90\\', \\'acr response at day 15: 0\\', \\'acr response at day 15: 50\\']}\\n'"]},"execution_count":139,"metadata":{},"output_type":"execute_result"}],"source":["f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n","1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n","2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n","   (1) Is there human data available for this trait?\n","   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n","   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n","   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n","\n","Background information about the dataset:\n","{background_info}\n","\n","Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n","{clinical_data_unique}\n","'''"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978240504Z","start_time":"2023-12-31T03:58:04.922365324Z"},"collapsed":false,"id":"WYm2ZV9ZiaD_"},"source":["Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n","Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."]},{"cell_type":"code","execution_count":148,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508295660Z","start_time":"2023-12-31T21:19:35.493015671Z"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704459453619,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"PT7NFhWRiaEA"},"outputs":[],"source":["is_gene_availabe = True\n","trait_row = 1\n","age_row = None\n","gender_row = None\n","\n","trait_type = 'binary'"]},{"cell_type":"code","execution_count":149,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508365153Z","start_time":"2023-12-31T21:19:35.493151450Z"},"executionInfo":{"elapsed":457,"status":"ok","timestamp":1704459467200,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"gA7w5UyliaEA"},"outputs":[],"source":["is_available = is_gene_availabe and (trait_row is not None)\n","if not is_available:\n","    save_cohort_info(cohort, JSON_PATH, is_available)\n","    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"]},{"cell_type":"code","execution_count":166,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508426069Z","start_time":"2023-12-31T21:19:35.493220101Z"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704459469360,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"tY1MZ2cMiaEA"},"outputs":[],"source":["\n","\n","def convert_trait(value):\n","    if 'Healthy' in value:\n","        return 0\n","    elif  'SJIA' in value:\n","        return 1\n","    else:\n","        return None  # Unknown value\n","\n","\n","\n","def convert_age(value):\n","    try:\n","        return float(value)  # Assuming age is a numeric value\n","    except ValueError:\n","        return None  # Unknown or invalid value\n","\n","def convert_gender(value):\n","    if value in ['male', 'female']:\n","        return value\n","    else:\n","        return None  # Unknown or invalid value\n","\n","\n","\n","\n","\n"]},{"cell_type":"code","execution_count":167,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GSM2111993</th>\n","      <th>GSM2111994</th>\n","      <th>GSM2111995</th>\n","      <th>GSM2111996</th>\n","      <th>GSM2111997</th>\n","      <th>GSM2111998</th>\n","      <th>GSM2111999</th>\n","      <th>GSM2112000</th>\n","      <th>GSM2112001</th>\n","      <th>GSM2112002</th>\n","      <th>...</th>\n","      <th>GSM2112189</th>\n","      <th>GSM2112190</th>\n","      <th>GSM2112191</th>\n","      <th>GSM2112192</th>\n","      <th>GSM2112193</th>\n","      <th>GSM2112194</th>\n","      <th>GSM2112195</th>\n","      <th>GSM2112196</th>\n","      <th>GSM2112197</th>\n","      <th>GSM2112198</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>Autoinflammatory Disorders</th>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1 rows × 206 columns</p>\n","</div>"],"text/plain":["                            GSM2111993  GSM2111994  GSM2111995  GSM2111996  \\\n","Autoinflammatory Disorders           1           1           1           1   \n","\n","                            GSM2111997  GSM2111998  GSM2111999  GSM2112000  \\\n","Autoinflammatory Disorders           1           1           1           1   \n","\n","                            GSM2112001  GSM2112002  ...  GSM2112189  \\\n","Autoinflammatory Disorders           1           1  ...           0   \n","\n","                            GSM2112190  GSM2112191  GSM2112192  GSM2112193  \\\n","Autoinflammatory Disorders           0           0           0           0   \n","\n","                            GSM2112194  GSM2112195  GSM2112196  GSM2112197  \\\n","Autoinflammatory Disorders           1           1           1           1   \n","\n","                            GSM2112198  \n","Autoinflammatory Disorders           1  \n","\n","[1 rows x 206 columns]"]},"execution_count":167,"metadata":{},"output_type":"execute_result"}],"source":["selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n","                                                      convert_age=convert_age, gender_row=gender_row,\n","                                                      convert_gender=convert_gender)\n","selected_clinical_data.head()"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978664193Z","start_time":"2023-12-31T03:58:04.966117261Z"},"collapsed":false,"id":"Avmg2XvQiaEA"},"source":["### Genetic data preprocessing and final filtering"]},{"cell_type":"code","execution_count":168,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590261560Z","start_time":"2023-12-31T21:19:35.493320362Z"},"colab":{"base_uri":"https://localhost:8080/","height":256},"executionInfo":{"elapsed":2540,"status":"ok","timestamp":1704459490752,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"VnaNkDORiaEA","outputId":"fc8c9388-a369-4c8e-a945-9bfedde0f51f"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GSM2111993</th>\n","      <th>GSM2111994</th>\n","      <th>GSM2111995</th>\n","      <th>GSM2111996</th>\n","      <th>GSM2111997</th>\n","      <th>GSM2111998</th>\n","      <th>GSM2111999</th>\n","      <th>GSM2112000</th>\n","      <th>GSM2112001</th>\n","      <th>GSM2112002</th>\n","      <th>...</th>\n","      <th>GSM2112189</th>\n","      <th>GSM2112190</th>\n","      <th>GSM2112191</th>\n","      <th>GSM2112192</th>\n","      <th>GSM2112193</th>\n","      <th>GSM2112194</th>\n","      <th>GSM2112195</th>\n","      <th>GSM2112196</th>\n","      <th>GSM2112197</th>\n","      <th>GSM2112198</th>\n","    </tr>\n","    <tr>\n","      <th>ID</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>1007_s_at</th>\n","      <td>37.73324</td>\n","      <td>35.302510</td>\n","      <td>40.570510</td>\n","      <td>23.247160</td>\n","      <td>44.120880</td>\n","      <td>47.03959</td>\n","      <td>33.27792</td>\n","      <td>62.789380</td>\n","      <td>43.628440</td>\n","      <td>43.850280</td>\n","      <td>...</td>\n","      <td>102.928140</td>\n","      <td>108.296920</td>\n","      <td>56.44268</td>\n","      <td>93.483740</td>\n","      <td>66.957990</td>\n","      <td>33.932170</td>\n","      <td>45.005060</td>\n","      <td>33.083060</td>\n","      <td>42.862820</td>\n","      <td>80.554360</td>\n","    </tr>\n","    <tr>\n","      <th>1053_at</th>\n","      <td>119.09356</td>\n","      <td>56.774720</td>\n","      <td>87.840060</td>\n","      <td>93.492870</td>\n","      <td>63.508300</td>\n","      <td>70.96652</td>\n","      <td>95.39125</td>\n","      <td>60.386020</td>\n","      <td>77.106000</td>\n","      <td>67.987640</td>\n","      <td>...</td>\n","      <td>159.680080</td>\n","      <td>338.087600</td>\n","      <td>177.09520</td>\n","      <td>184.002390</td>\n","      <td>151.065070</td>\n","      <td>219.835440</td>\n","      <td>200.553650</td>\n","      <td>187.176430</td>\n","      <td>209.607070</td>\n","      <td>164.314760</td>\n","    </tr>\n","    <tr>\n","      <th>117_at</th>\n","      <td>646.04120</td>\n","      <td>401.971100</td>\n","      <td>788.235300</td>\n","      <td>657.719400</td>\n","      <td>521.005100</td>\n","      <td>929.27900</td>\n","      <td>794.13030</td>\n","      <td>132.824500</td>\n","      <td>658.225300</td>\n","      <td>554.553200</td>\n","      <td>...</td>\n","      <td>971.029100</td>\n","      <td>1331.777200</td>\n","      <td>1090.81180</td>\n","      <td>1006.189500</td>\n","      <td>638.855200</td>\n","      <td>1656.521800</td>\n","      <td>1028.245900</td>\n","      <td>869.116700</td>\n","      <td>1202.487100</td>\n","      <td>467.813600</td>\n","    </tr>\n","    <tr>\n","      <th>121_at</th>\n","      <td>58.07843</td>\n","      <td>31.556470</td>\n","      <td>46.542570</td>\n","      <td>34.061020</td>\n","      <td>48.228420</td>\n","      <td>52.29905</td>\n","      <td>42.85798</td>\n","      <td>40.027200</td>\n","      <td>39.571780</td>\n","      <td>41.547660</td>\n","      <td>...</td>\n","      <td>42.663370</td>\n","      <td>64.750920</td>\n","      <td>39.09132</td>\n","      <td>40.819100</td>\n","      <td>49.717720</td>\n","      <td>64.835870</td>\n","      <td>40.161670</td>\n","      <td>42.936320</td>\n","      <td>49.000850</td>\n","      <td>49.220060</td>\n","    </tr>\n","    <tr>\n","      <th>1255_g_at</th>\n","      <td>4.13666</td>\n","      <td>2.479351</td>\n","      <td>3.379874</td>\n","      <td>3.134788</td>\n","      <td>2.736602</td>\n","      <td>3.56498</td>\n","      <td>2.92125</td>\n","      <td>2.637146</td>\n","      <td>2.838363</td>\n","      <td>2.957969</td>\n","      <td>...</td>\n","      <td>2.299729</td>\n","      <td>3.113178</td>\n","      <td>2.88688</td>\n","      <td>2.936406</td>\n","      <td>2.661781</td>\n","      <td>2.422376</td>\n","      <td>2.855924</td>\n","      <td>2.330613</td>\n","      <td>3.009756</td>\n","      <td>2.264879</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 206 columns</p>\n","</div>"],"text/plain":["           GSM2111993  GSM2111994  GSM2111995  GSM2111996  GSM2111997  \\\n","ID                                                                      \n","1007_s_at    37.73324   35.302510   40.570510   23.247160   44.120880   \n","1053_at     119.09356   56.774720   87.840060   93.492870   63.508300   \n","117_at      646.04120  401.971100  788.235300  657.719400  521.005100   \n","121_at       58.07843   31.556470   46.542570   34.061020   48.228420   \n","1255_g_at     4.13666    2.479351    3.379874    3.134788    2.736602   \n","\n","           GSM2111998  GSM2111999  GSM2112000  GSM2112001  GSM2112002  ...  \\\n","ID                                                                     ...   \n","1007_s_at    47.03959    33.27792   62.789380   43.628440   43.850280  ...   \n","1053_at      70.96652    95.39125   60.386020   77.106000   67.987640  ...   \n","117_at      929.27900   794.13030  132.824500  658.225300  554.553200  ...   \n","121_at       52.29905    42.85798   40.027200   39.571780   41.547660  ...   \n","1255_g_at     3.56498     2.92125    2.637146    2.838363    2.957969  ...   \n","\n","           GSM2112189   GSM2112190  GSM2112191   GSM2112192  GSM2112193  \\\n","ID                                                                        \n","1007_s_at  102.928140   108.296920    56.44268    93.483740   66.957990   \n","1053_at    159.680080   338.087600   177.09520   184.002390  151.065070   \n","117_at     971.029100  1331.777200  1090.81180  1006.189500  638.855200   \n","121_at      42.663370    64.750920    39.09132    40.819100   49.717720   \n","1255_g_at    2.299729     3.113178     2.88688     2.936406    2.661781   \n","\n","            GSM2112194   GSM2112195  GSM2112196   GSM2112197  GSM2112198  \n","ID                                                                        \n","1007_s_at    33.932170    45.005060   33.083060    42.862820   80.554360  \n","1053_at     219.835440   200.553650  187.176430   209.607070  164.314760  \n","117_at     1656.521800  1028.245900  869.116700  1202.487100  467.813600  \n","121_at       64.835870    40.161670   42.936320    49.000850   49.220060  \n","1255_g_at     2.422376     2.855924    2.330613     3.009756    2.264879  \n","\n","[5 rows x 206 columns]"]},"execution_count":168,"metadata":{},"output_type":"execute_result"}],"source":["genetic_data = get_genetic_data(matrix_file)\n","genetic_data.head()"]},{"cell_type":"code","execution_count":169,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590443507Z","start_time":"2023-12-31T21:19:35.536713559Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704459492414,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"aCWv6oYZiaEA","outputId":"3f614e5a-04c2-4972-8dbf-af619c0254f5"},"outputs":[{"data":{"text/plain":["['1007_s_at',\n"," '1053_at',\n"," '117_at',\n"," '121_at',\n"," '1255_g_at',\n"," '1294_at',\n"," '1316_at',\n"," '1320_at',\n"," '1405_i_at',\n"," '1431_at',\n"," '1438_at',\n"," '1487_at',\n"," '1494_f_at',\n"," '1552256_a_at',\n"," '1552257_a_at',\n"," '1552258_at',\n"," '1552261_at',\n"," '1552263_at',\n"," '1552264_a_at',\n"," '1552266_at']"]},"execution_count":169,"metadata":{},"output_type":"execute_result"}],"source":["gene_row_ids = genetic_data.index[:20].tolist()\n","gene_row_ids"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:30:41.595335164Z","start_time":"2023-12-31T03:30:41.513232329Z"},"collapsed":false,"id":"_GDJH2wbiaEA"},"source":["Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"-oGa4Zz0iaEA"},"source":["Reference prompt:"]},{"cell_type":"code","execution_count":170,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590590427Z","start_time":"2023-12-31T21:19:35.536780517Z"},"colab":{"base_uri":"https://localhost:8080/","height":125},"executionInfo":{"elapsed":515,"status":"ok","timestamp":1704459495363,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"icIqZ_XaiaEB","outputId":"88837551-d222-43d9-8507-a2febc0aee22"},"outputs":[{"data":{"text/plain":["\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1552256_a_at', '1552257_a_at', '1552258_at', '1552261_at', '1552263_at', '1552264_a_at', '1552266_at']\\n\""]},"execution_count":170,"metadata":{},"output_type":"execute_result"}],"source":["f'''\n","Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n","requires_gene_mapping = (True or False)\n","\n","Row headers:\n","{gene_row_ids}\n","'''"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"lqpef-AEiaEB"},"source":["If not required, jump directly to the gene normalization step"]},{"cell_type":"code","execution_count":171,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590648558Z","start_time":"2023-12-31T21:19:35.536821375Z"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704459497156,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"mjjOtDejiaEB"},"outputs":[],"source":["requires_gene_mapping = True"]},{"cell_type":"code","execution_count":172,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590695457Z","start_time":"2023-12-31T21:19:35.536910836Z"},"executionInfo":{"elapsed":41476,"status":"ok","timestamp":1704459540538,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"ddUHkPQeiaEB"},"outputs":[{"name":"stdout","output_type":"stream","text":["{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n"]}],"source":["if requires_gene_mapping:\n","    gene_annotation = get_gene_annotation(soft_file)\n","    gene_annotation_summary = preview_df(gene_annotation)\n","    print(gene_annotation_summary)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978931927Z","start_time":"2023-12-31T03:58:04.966328339Z"},"collapsed":false,"id":"iX79MmEKiaEB"},"source":["Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n","Reference prompt:"]},{"cell_type":"code","execution_count":173,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590742236Z","start_time":"2023-12-31T21:19:35.537050893Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":416,"status":"ok","timestamp":1704459545733,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"IFcuQ4S9iaEB","outputId":"003b7f79-3587-425c-a76f-364f7668bcf5"},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n","    \n"]}],"source":["if requires_gene_mapping:\n","    print(f'''\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {gene_annotation_summary}\n","    ''')"]},{"cell_type":"code","execution_count":174,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590786881Z","start_time":"2023-12-31T21:19:35.537132859Z"},"executionInfo":{"elapsed":3382,"status":"ok","timestamp":1704459600983,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"pCVgeE6DiaEB"},"outputs":[],"source":["if requires_gene_mapping:\n","    probe_id_key = 'ID'\n","    gene_symb_key = 'Gene Symbol'\n","    gene_mapping = get_gene_mapping(gene_annotation, probe_id_key, gene_symb_key)\n","    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"]},{"cell_type":"code","execution_count":175,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590837318Z","start_time":"2023-12-31T21:19:35.537209586Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":79400,"status":"ok","timestamp":1704459682112,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"aqJRhdbFiaEB","outputId":"e32013d3-5c4b-4dd3-ec63-603a958dd08d"},"outputs":[],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"]},{"cell_type":"code","execution_count":75,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590902191Z","start_time":"2023-12-31T21:19:35.537285901Z"},"executionInfo":{"elapsed":410,"status":"ok","timestamp":1704459686452,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"istECoP1iaEB"},"outputs":[],"source":["merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n","# The preprocessing runs through, which means is_available should be True\n","is_available = True"]},{"cell_type":"code","execution_count":76,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.591095170Z","start_time":"2023-12-31T21:19:35.538708445Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704459689000,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"tzrpD6LciaEB","outputId":"ee927911-f436-4e12-91b7-85e031ccd63b"},"outputs":[{"name":"stdout","output_type":"stream","text":["The merged dataset contains 206 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"]},{"cell_type":"code","execution_count":77,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610849182Z","start_time":"2023-12-31T21:19:35.541210258Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704459690597,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"opVFXzd4iaEC","outputId":"aeee804a-52e1-4699-bcc4-e074b2b9a2ee"},"outputs":[{"name":"stdout","output_type":"stream","text":["For the feature 'Autoinflammatory Disorders', the least common label is '0.0' with 22 occurrences. This represents 10.68% of the dataset.\n","The distribution of the feature 'Autoinflammatory Disorders' in this dataset is fine.\n","\n"]},{"data":{"text/plain":["False"]},"execution_count":77,"metadata":{},"output_type":"execute_result"}],"source":["is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"]},{"cell_type":"code","execution_count":78,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610923234Z","start_time":"2023-12-31T21:19:35.584848473Z"},"executionInfo":{"elapsed":949,"status":"ok","timestamp":1704459693473,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"wFORrTwciaEC"},"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"]},{"cell_type":"code","execution_count":79,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610970734Z","start_time":"2023-12-31T21:19:35.585042694Z"},"executionInfo":{"elapsed":5441,"status":"ok","timestamp":1704459700767,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"EG3DaeSJiaEC"},"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-12-31T03:58:22.271683755Z","start_time":"2023-12-31T03:58:22.246557674Z"},"id":"-MTPhRGxJV7I"},"source":["### 3. Do regression & Cross Validation"]},{"cell_type":"code","execution_count":28,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.459442087Z","start_time":"2023-12-31T21:19:44.416219648Z"},"executionInfo":{"elapsed":533,"status":"ok","timestamp":1704459733416,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"},"user_tz":-480},"id":"v-dUlz9_iaEC"},"outputs":[],"source":["best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH, 'has_age')\n","best_cohort"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.808945960Z","start_time":"2023-12-31T21:19:44.799467072Z"},"id":"r15ErhKhiaEC","outputId":"cf26557f-7ea5-4980-d4f5-e77fe72b50da"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>cohort_id</th>\n","      <th>is_usable</th>\n","      <th>is_available</th>\n","      <th>is_biased</th>\n","      <th>has_age</th>\n","      <th>has_gender</th>\n","      <th>sample_size</th>\n","      <th>note</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Xena</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>1215.0</td>\n","      <td></td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>GSE248830</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>38.0</td>\n","      <td></td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["   cohort_id  is_usable  is_available is_biased has_age has_gender  \\\n","0       Xena       True          True     False    True       True   \n","3  GSE248830       True          True     False    True       True   \n","\n","   sample_size note  \n","0       1215.0       \n","3         38.0       "]},"execution_count":45,"metadata":{},"output_type":"execute_result"}],"source":["ranked_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.222719978Z","start_time":"2023-12-31T21:19:45.162873006Z"},"id":"2c15dYzIiaEC"},"outputs":[],"source":["merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, best_cohort+'.csv'))"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.234945878Z","start_time":"2023-12-31T21:19:52.231604618Z"},"id":"LOAj-G92iaEC","outputId":"519f4ae4-7c39-4ba0-cbb4-f2d4f67dcc44"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Breast Cancer</th>\n","      <th>Age</th>\n","      <th>Gender</th>\n","      <th>ARHGEF10L</th>\n","      <th>HIF3A</th>\n","      <th>RNF17</th>\n","      <th>RNF10</th>\n","      <th>RNF11</th>\n","      <th>RNF13</th>\n","      <th>GTF2IP1</th>\n","      <th>...</th>\n","      <th>TULP2</th>\n","      <th>NPY5R</th>\n","      <th>GNGT2</th>\n","      <th>GNGT1</th>\n","      <th>TULP3</th>\n","      <th>PTRF</th>\n","      <th>BCL6B</th>\n","      <th>GSTK1</th>\n","      <th>SELP</th>\n","      <th>SELS</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>1</td>\n","      <td>55.0</td>\n","      <td>0.0</td>\n","      <td>0.607308</td>\n","      <td>-3.194126</td>\n","      <td>-0.531035</td>\n","      <td>-0.145872</td>\n","      <td>0.237422</td>\n","      <td>-0.29921</td>\n","      <td>-0.142694</td>\n","      <td>...</td>\n","      <td>-0.748878</td>\n","      <td>0.565583</td>\n","      <td>-0.767233</td>\n","      <td>-1.28139</td>\n","      <td>-0.271377</td>\n","      <td>-0.492286</td>\n","      <td>0.360373</td>\n","      <td>1.067905</td>\n","      <td>0.076267</td>\n","      <td>-0.392212</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1</td>\n","      <td>50.0</td>\n","      <td>0.0</td>\n","      <td>-0.641192</td>\n","      <td>-4.928226</td>\n","      <td>0.095465</td>\n","      <td>0.098128</td>\n","      <td>-0.541978</td>\n","      <td>-0.32291</td>\n","      <td>-0.044694</td>\n","      <td>...</td>\n","      <td>0.312922</td>\n","      <td>0.079683</td>\n","      <td>0.475267</td>\n","      <td>1.99661</td>\n","      <td>0.161423</td>\n","      <td>0.273714</td>\n","      <td>0.644673</td>\n","      <td>-0.428695</td>\n","      <td>0.068667</td>\n","      <td>-0.043812</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>1</td>\n","      <td>62.0</td>\n","      <td>0.0</td>\n","      <td>1.082808</td>\n","      <td>-4.623726</td>\n","      <td>-0.531035</td>\n","      <td>0.484028</td>\n","      <td>-0.183678</td>\n","      <td>-0.91901</td>\n","      <td>0.261106</td>\n","      <td>...</td>\n","      <td>-0.748878</td>\n","      <td>-0.656117</td>\n","      <td>-0.216733</td>\n","      <td>0.21081</td>\n","      <td>0.101023</td>\n","      <td>0.660514</td>\n","      <td>1.295073</td>\n","      <td>0.915105</td>\n","      <td>0.168567</td>\n","      <td>0.047788</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>1</td>\n","      <td>52.0</td>\n","      <td>0.0</td>\n","      <td>0.121608</td>\n","      <td>-2.881526</td>\n","      <td>-0.531035</td>\n","      <td>0.179128</td>\n","      <td>0.039222</td>\n","      <td>-0.45491</td>\n","      <td>0.180306</td>\n","      <td>...</td>\n","      <td>0.415422</td>\n","      <td>-0.178417</td>\n","      <td>-0.211233</td>\n","      <td>-1.28139</td>\n","      <td>0.108023</td>\n","      <td>1.132814</td>\n","      <td>0.496773</td>\n","      <td>0.240105</td>\n","      <td>3.099767</td>\n","      <td>0.112888</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>1</td>\n","      <td>50.0</td>\n","      <td>0.0</td>\n","      <td>0.420208</td>\n","      <td>-3.282726</td>\n","      <td>-0.531035</td>\n","      <td>-0.020972</td>\n","      <td>-0.117978</td>\n","      <td>-0.55781</td>\n","      <td>-0.173794</td>\n","      <td>...</td>\n","      <td>0.685222</td>\n","      <td>-0.698717</td>\n","      <td>-0.484233</td>\n","      <td>-1.28139</td>\n","      <td>-0.223577</td>\n","      <td>1.222714</td>\n","      <td>0.225573</td>\n","      <td>-0.301995</td>\n","      <td>0.200067</td>\n","      <td>-0.080212</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 20533 columns</p>\n","</div>"],"text/plain":["   Breast Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17     RNF10  \\\n","0              1  55.0     0.0   0.607308 -3.194126 -0.531035 -0.145872   \n","1              1  50.0     0.0  -0.641192 -4.928226  0.095465  0.098128   \n","2              1  62.0     0.0   1.082808 -4.623726 -0.531035  0.484028   \n","3              1  52.0     0.0   0.121608 -2.881526 -0.531035  0.179128   \n","4              1  50.0     0.0   0.420208 -3.282726 -0.531035 -0.020972   \n","\n","      RNF11    RNF13   GTF2IP1  ...     TULP2     NPY5R     GNGT2    GNGT1  \\\n","0  0.237422 -0.29921 -0.142694  ... -0.748878  0.565583 -0.767233 -1.28139   \n","1 -0.541978 -0.32291 -0.044694  ...  0.312922  0.079683  0.475267  1.99661   \n","2 -0.183678 -0.91901  0.261106  ... -0.748878 -0.656117 -0.216733  0.21081   \n","3  0.039222 -0.45491  0.180306  ...  0.415422 -0.178417 -0.211233 -1.28139   \n","4 -0.117978 -0.55781 -0.173794  ...  0.685222 -0.698717 -0.484233 -1.28139   \n","\n","      TULP3      PTRF     BCL6B     GSTK1      SELP      SELS  \n","0 -0.271377 -0.492286  0.360373  1.067905  0.076267 -0.392212  \n","1  0.161423  0.273714  0.644673 -0.428695  0.068667 -0.043812  \n","2  0.101023  0.660514  1.295073  0.915105  0.168567  0.047788  \n","3  0.108023  1.132814  0.496773  0.240105  3.099767  0.112888  \n","4 -0.223577  1.222714  0.225573 -0.301995  0.200067 -0.080212  \n","\n","[5 rows x 20533 columns]"]},"execution_count":47,"metadata":{},"output_type":"execute_result"}],"source":["merged_data.head()"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.312540150Z","start_time":"2023-12-31T21:19:52.235772334Z"},"id":"vtzuVZpLiaEC"},"outputs":[],"source":["# If both age and gender features are available, select 'age' as the condition.\n","condition = 'Age'\n","# Remove the other condition to prevent interference.\n","merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.323468266Z","start_time":"2023-12-31T21:19:52.290903989Z"},"id":"LJDde0WQiaEC"},"outputs":[],"source":["X = merged_data.drop(columns=[TRAIT, condition]).values\n","Y = merged_data[TRAIT].values\n","Z = merged_data[condition].values"]},{"cell_type":"markdown","metadata":{"collapsed":false,"id":"VKBCcPkGiaEC"},"source":["Select the appropriate regression model depending on whether the dataset shows batch effect."]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.930840782Z","start_time":"2023-12-31T21:19:52.319092572Z"},"id":"0mp8bOqciaED","outputId":"2e542440-11b5-4751-9d72-460b69859c8f"},"outputs":[{"data":{"text/plain":["True"]},"execution_count":50,"metadata":{},"output_type":"execute_result"}],"source":["has_batch_effect = detect_batch_effect(X)\n","has_batch_effect"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.931029722Z","start_time":"2023-12-31T21:20:00.929710828Z"},"id":"IITCrXFGiaED"},"outputs":[],"source":["if has_batch_effect:\n","    model_constructor = VariableSelection\n","    model_params = {}\n","else:\n","    model_constructor = Lasso\n","    model_params = {'alpha': 1.0, 'random_state': 42}"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:20.709459824Z","start_time":"2023-12-31T21:20:00.931657949Z"},"id":"qWZ4p5LRiaED","outputId":"94d8262b-dfc7-44fc-8a36-b36a140bcd94"},"outputs":[{"name":"stderr","output_type":"stream","text":["/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n"]},{"name":"stdout","output_type":"stream","text":["The cross-validation accuracy is 91.28% ± 1.50%\n"]}],"source":["cv_mean, cv_std = cross_validation(X, Y, Z, model_constructor, model_params, target_type=trait_type)"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:24.980870159Z","start_time":"2023-12-31T21:20:20.516710807Z"},"id":"6PpgQE0kiaED","outputId":"0466e0eb-88f5-4e9a-8236-666d416f1b1d"},"outputs":[{"name":"stderr","output_type":"stream","text":["/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n"]}],"source":["# Train regression model on the whole dataset to identify significant genes\n","model = ResidualizationRegressor(model_constructor, model_params)\n","normalized_X, _ = normalize_data(X)\n","normalized_Z, _ = normalize_data(Z)\n","model.fit(normalized_X, Y, normalized_Z)"]},{"cell_type":"markdown","metadata":{"ExecuteTime":{"end_time":"2023-10-14T04:50:59.072075075Z","start_time":"2023-10-14T04:50:38.739499998Z"},"id":"EjJrxbvb4nlj"},"source":["### 5. Discussion and report"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:25.000830410Z","start_time":"2023-12-31T21:20:24.982834541Z"},"id":"0VidWNDkiaED","outputId":"ae1dfede-e05a-40bf-80ca-203e5fffcef5"},"outputs":[{"name":"stdout","output_type":"stream","text":["Effect of the condition on the target variable:\n","Variable: Age\n","Coefficient: 0.0065\n","p-value: nan\n","\n","Found 76 significant genes associated with the trait 'Breast Cancer' conditional on the factor 'Age', with corrected p-value < 0.05:\n","    Variable  Coefficient  corrected_p_value\n","     COL10A1     0.136339       3.404090e-27\n","         BGN     0.095422       1.004817e-14\n","    PPAPDC1A     0.081296       9.785054e-13\n","       MMP11     0.080354       1.123019e-10\n","       WISP1     0.074888       9.613830e-10\n","     HSD17B6     0.051940       6.980811e-08\n","      CTHRC1     0.058310       2.210553e-07\n","       MFAP5    -0.054250       2.706783e-07\n","      CLEC5A     0.047846       8.877914e-07\n","   C20orf103     0.048957       8.877914e-07\n","          GC     0.027375       1.419502e-05\n","      OR51D1     0.022908       1.644207e-05\n","        COMP     0.041129       4.366228e-05\n","LOC100169752     0.019398       4.366228e-05\n","      LRRC15     0.052746       4.494130e-05\n","         FN1     0.063527       5.027562e-05\n","     C1QTNF6     0.049130       1.220626e-04\n","       INHBA     0.052621       1.384526e-04\n","       FNDC1     0.044080       1.724540e-04\n","        VCAN     0.062919       3.541404e-04\n","   LOC283867     0.039438       3.541404e-04\n","     SLC24A2     0.045359       3.929641e-04\n","        ECM2    -0.066303       4.895350e-04\n","       CILP2     0.040613       4.895350e-04\n","         DCN    -0.062282       5.405761e-04\n","        HEYL     0.039376       6.133304e-04\n","         FAP     0.062297       6.903458e-04\n","     OLFML2A    -0.037988       8.815285e-04\n","       MFAP2     0.047162       8.815285e-04\n","       CCRL1    -0.036802       8.815285e-04\n","       CD248    -0.046771       8.815285e-04\n","       P4HA3     0.044037       8.815285e-04\n","     C5orf13     0.037746       1.590045e-03\n","      COL5A2     0.064118       2.057973e-03\n","         F2R     0.041164       2.101197e-03\n","      OLFML1    -0.061570       3.573228e-03\n","       PRDM8    -0.040662       3.786028e-03\n","   POM121L12     0.028947       3.941936e-03\n","       LOXL1     0.044681       4.547038e-03\n","       SPARC     0.053330       5.391493e-03\n","      LEFTY2     0.027338       5.753501e-03\n","    SERPINE1     0.028865       5.753501e-03\n","       OR2L3    -0.016537       6.402547e-03\n","     CYP19A1     0.022029       6.651093e-03\n","         MAF    -0.038952       7.098913e-03\n","       GNG11    -0.068155       1.242228e-02\n","       CFHR5    -0.018470       1.366939e-02\n","      CLEC3B    -0.043039       1.366939e-02\n","        NRG3    -0.019375       1.366939e-02\n","      COL1A1     0.046342       1.427700e-02\n","      PCOLCE    -0.049663       1.836229e-02\n","     TMEM119     0.037238       1.922482e-02\n","       KCNC1    -0.024520       2.248216e-02\n","       WNT8A     0.014598       2.592754e-02\n","        RECK    -0.052887       2.781665e-02\n","        SNCB     0.018240       3.140935e-02\n","      KBTBD6    -0.029237       3.140935e-02\n","       SRPX2    -0.034794       3.140935e-02\n","     OLFML2B     0.036407       3.214194e-02\n","      PLXDC1    -0.032863       3.219694e-02\n","      ADAM12     0.040368       3.309987e-02\n","      GRIN2A    -0.021383       3.340752e-02\n","      COL5A1     0.050110       3.456174e-02\n","      USHBP1    -0.050144       3.618703e-02\n","       CYLC2    -0.016414       3.622489e-02\n","     TMEM90B     0.034482       3.622489e-02\n","      RESP18    -0.014081       3.622489e-02\n","        RGS4     0.022820       3.855158e-02\n","        CNTF    -0.020244       3.880001e-02\n","    ADAMTSL2     0.030360       4.155675e-02\n","        CST2     0.019834       4.218502e-02\n","       GRID1    -0.028724       4.218502e-02\n","       MATN3     0.028706       4.316484e-02\n","       GGT8P    -0.017000       4.494972e-02\n","      GPR101     0.014064       4.913174e-02\n","      KIF26B     0.031349       4.960098e-02\n"]},{"name":"stderr","output_type":"stream","text":["/home/techt/AI4Science_gold_standard/stage1/Haoyang/utils.py:492: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"]}],"source":["feature_cols = merged_data.columns.tolist()\n","feature_cols.remove(TRAIT)\n","\n","if has_batch_effect:\n","    report_result_from_lmm(model, feature_cols, TRAIT, condition, threshold=0.05, save_output=True,\n","                           output_dir=OUTPUT_DIR)\n","else:\n","    report_result_from_lasso(model, feature_cols, TRAIT, condition, save_output=True, output_dir=OUTPUT_DIR)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bQbxWUPBiaED"},"outputs":[],"source":[]}],"metadata":{"colab":{"machine_shape":"hm","provenance":[],"toc_visible":true},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.9"}},"nbformat":4,"nbformat_minor":0}
