{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"toc_visible":true,"machine_shape":"hm"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Gold standard curation: Preprocessing and single-step regression"],"metadata":{"collapsed":false,"id":"8Q7_HywEiaDx"}},{"cell_type":"markdown","source":["In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n","\n","1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n","2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"],"metadata":{"collapsed":false,"id":"pCI2xoVwiaDy"}},{"cell_type":"markdown","source":["# 1. Basic setup"],"metadata":{"collapsed":false,"id":"3lP_4rV0iaDz"}},{"cell_type":"code","execution_count":2,"outputs":[],"source":["import os\n","\n","# Set your preferred name\n","USER = \"Shuyi\"\n","# Set the data and output directories\n","DATA_ROOT = '/content/drive/MyDrive/AI4Science_Public/data'\n","OUTPUT_ROOT = '/content/drive/MyDrive/AI4Science_Public/output1'\n","TRAIT = \"Autoinflammatory Disorders\"\n","\n","OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n","JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n","if not os.path.exists(OUTPUT_DIR):\n","    os.makedirs(OUTPUT_DIR, exist_ok=True)\n","\n","# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n","NORMALIZE_GENE = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:07.313098989Z","start_time":"2023-12-31T21:19:07.312718292Z"},"id":"LqmCfkN7iaDz","executionInfo":{"status":"ok","timestamp":1704459407389,"user_tz":-480,"elapsed":1480,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":1,"metadata":{"id":"jMTdsU43vVg3","ExecuteTime":{"end_time":"2023-12-31T21:19:07.680659936Z","start_time":"2023-12-31T21:19:07.666959145Z"},"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459380373,"user_tz":-480,"elapsed":23046,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"eb82b8af-3908-4a22-f730-a847d9b7ffca"},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}],"source":["# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n","\n","import os\n","from google.colab import drive\n","\n","drive.mount('/content/drive', force_remount=True)\n","proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n","os.chdir(proj_dir)"]},{"cell_type":"code","source":["import sys\n","sys.path.append('/content/drive/MyDrive/Colab Notebooks')\n","!pip install mygene\n","!pip install sparse_lmm"],"metadata":{"id":"T1jKdOOHscwu","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459420000,"user_tz":-480,"elapsed":11054,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"b7e81b10-ce5b-4e28-c10d-e30f9e83807d"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting mygene\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Collecting biothings-client>=0.2.6 (from mygene)\n","  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n","Requirement already satisfied: requests>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from biothings-client>=0.2.6->mygene) (2.31.0)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2023.11.17)\n","Installing collected packages: biothings-client, mygene\n","Successfully installed biothings-client-0.3.1 mygene-3.2.2\n","Collecting sparse_lmm\n","  Downloading sparse_lmm-0.4-py3-none-any.whl (7.1 kB)\n","Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from sparse_lmm) (1.11.4)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from sparse_lmm) (1.23.5)\n","Installing collected packages: sparse_lmm\n","Successfully installed sparse_lmm-0.4\n"]}]},{"cell_type":"markdown","source":["# 2. Data preprocessing and selection"],"metadata":{"collapsed":false,"id":"iNY8dzJAiaD0"}},{"cell_type":"markdown","source":["## 2.1. The TCGA Xena dataset"],"metadata":{"collapsed":false,"id":"7aNtx3jjiaD1"}},{"cell_type":"markdown","source":["In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."],"metadata":{"collapsed":false,"id":"nn68pLQ1iaD1"}},{"cell_type":"code","execution_count":4,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['.DS_Store',\n"," 'TCGA_Esophageal_Cancer_(ESCA)',\n"," 'TCGA_Pheochromocytoma_Paraganglioma_(PCPG)',\n"," 'TCGA_Bile_Duct_Cancer_(CHOL)',\n"," 'TCGA_Kidney_Chromophobe_(KICH)',\n"," 'TCGA_Thyroid_Cancer_(THCA)',\n"," 'TCGA_Sarcoma_(SARC)',\n"," 'TCGA_Adrenocortical_Cancer_(ACC)',\n"," 'TCGA_Thymoma_(THYM)',\n"," 'TCGA_Glioblastoma_(GBM)']"]},"metadata":{},"execution_count":4}],"source":["import os\n","from utils import *\n","\n","dataset = 'TCGA'\n","dataset_dir = os.path.join(DATA_ROOT, dataset)\n","os.listdir(dataset_dir)[:10]"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.770095084Z","start_time":"2023-12-31T21:19:09.102451553Z"},"id":"hCwvC9PNiaD1","outputId":"9195a4d8-b7ac-4910-d90a-1a3bddba33d9","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459428432,"user_tz":-480,"elapsed":3886,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["If no match is found, jump directly to GEO in Part 2.2"],"metadata":{"collapsed":false,"id":"sksNMtjOiaD1"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["trait_subdir = \"TCGA_Breast_Cancer_(BRCA)\"\n","cohort = 'Xena'\n","# All the cancer traits in Xena are binary\n","trait_type = 'binary'\n","# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n","is_available = True\n","\n","clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_BRCA_clinicalMatrix')\n","genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.BRCA.sampleMap_HiSeqV2_PANCAN.gz')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.918243934Z","start_time":"2023-12-31T21:19:09.909021135Z"},"id":"hHU-R89oiaD1"}},{"cell_type":"code","source":["import pandas as pd\n","\n","clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n","genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n","age_col = gender_col = None"],"metadata":{"id":"MudwB-_iz7sc","ExecuteTime":{"end_time":"2023-12-31T21:19:13.613198409Z","start_time":"2023-12-31T21:19:10.409688003Z"}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"outputs":[],"source":["_, clinical_data_cols = check_rows_and_columns(clinical_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.617006167Z","start_time":"2023-12-31T21:19:13.614080653Z"},"id":"8XwEOygHiaD2"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"['AJCC_Stage_nature2012',\n 'Age_at_Initial_Pathologic_Diagnosis_nature2012',\n 'CN_Clusters_nature2012',\n 'Converted_Stage_nature2012',\n 'Days_to_Date_of_Last_Contact_nature2012',\n 'Days_to_date_of_Death_nature2012',\n 'ER_Status_nature2012',\n 'Gender_nature2012',\n 'HER2_Final_Status_nature2012',\n 'Integrated_Clusters_no_exp__nature2012']"},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["clinical_data_cols[:10]"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661538569Z","start_time":"2023-12-31T21:19:13.615946213Z"},"id":"TsoPD93diaD2","outputId":"b7b9aa38-4f63-4eea-8648-f208c80a4a60"}},{"cell_type":"markdown","source":["Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-28T20:20:11.841042813Z","start_time":"2023-12-28T20:20:11.834458206Z"},"id":"Ejzdxb0diaD2"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['AJCC_Stage_nature2012', 'Age_at_Initial_Pathologic_Diagnosis_nature2012', 'CN_Clusters_nature2012', 'Converted_Stage_nature2012', 'Days_to_Date_of_Last_Contact_nature2012', 'Days_to_date_of_Death_nature2012', 'ER_Status_nature2012', 'Gender_nature2012', 'HER2_Final_Status_nature2012', 'Integrated_Clusters_no_exp__nature2012', 'Integrated_Clusters_unsup_exp__nature2012', 'Integrated_Clusters_with_PAM50__nature2012', 'Metastasis_Coded_nature2012', 'Metastasis_nature2012', 'Node_Coded_nature2012', 'Node_nature2012', 'OS_Time_nature2012', 'OS_event_nature2012', 'PAM50Call_RNAseq', 'PAM50_mRNA_nature2012', 'PR_Status_nature2012', 'RPPA_Clusters_nature2012', 'SigClust_Intrinsic_mRNA_nature2012', 'SigClust_Unsupervised_mRNA_nature2012', 'Survival_Data_Form_nature2012', 'Tumor_T1_Coded_nature2012', 'Tumor_nature2012', 'Vital_Status_nature2012', '_INTEGRATION', '_PANCAN_CNA_PANCAN_K8', '_PANCAN_Cluster_Cluster_PANCAN', '_PANCAN_DNAMethyl_BRCA', '_PANCAN_DNAMethyl_PANCAN', '_PANCAN_RPPA_PANCAN_K8', '_PANCAN_UNC_RNAseq_PANCAN_K16', '_PANCAN_miRNA_PANCAN', '_PANCAN_mirna_BRCA', '_PANCAN_mutation_PANCAN', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_surgery_locoregional_procedure', 'additional_surgery_metastatic_procedure', 'age_at_initial_pathologic_diagnosis', 'anatomic_neoplasm_subdivision', 'axillary_lymph_node_stage_method_type', 'axillary_lymph_node_stage_other_method_descriptive_text', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'breast_cancer_surgery_margin_status', 'breast_carcinoma_estrogen_receptor_status', 'breast_carcinoma_immunohistochemistry_er_pos_finding_scale', 'breast_carcinoma_immunohistochemistry_pos_cell_score', 'breast_carcinoma_immunohistochemistry_prgstrn_rcptr_ps_fndng_scl', 'breast_carcinoma_primary_surgical_procedure_name', 'breast_carcinoma_progesterone_receptor_status', 'breast_carcinoma_surgical_procedure_name', 'breast_neoplasm_other_surgical_procedure_descriptive_text', 'cytokeratin_immunohistochemistry_staining_method_mcrmtstss_ndctr', 'days_to_additional_surgery_locoregional_procedure', 'days_to_additional_surgery_metastatic_procedure', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_last_known_alive', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'disease_code', 'distant_metastasis_present_ind2', 'er_detection_method_text', 'er_level_cell_percentage_category', 'fluorescence_in_st_hybrdztn_dgnstc_prcdr_chrmsm_17_sgnl_rslt_rng', 'followup_case_report_form_submission_reason', 'form_completion_date', 'gender', 'her2_and_centromere_17_positive_finding_other_measuremnt_scl_txt', 'her2_erbb_method_calculation_method_text', 'her2_erbb_pos_finding_cell_percent_category', 'her2_erbb_pos_finding_fluorescence_n_st_hybrdztn_clcltn_mthd_txt', 'her2_immunohistochemistry_level_result', 'her2_neu_and_centromere_17_copy_number_analysis_npt_ttl_nmbr_cnt', 'her2_neu_breast_carcinoma_copy_analysis_input_total_number', 'her2_neu_chromosone_17_signal_ratio_value', 'her2_neu_metastatic_breast_carcinoma_copy_analysis_inpt_ttl_nmbr', 'histological_type', 'histological_type_other', 'history_of_neoadjuvant_treatment', 'hr2_n_nd_cntrmr_17_cpy_nmbr_mtsttc_brst_crcnm_nlyss_npt_ttl_nmbr', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'immunohistochemistry_positive_cell_score', 'informed_consent_verified', 'init_pathology_dx_method_other', 'initial_pathologic_diagnosis_method', 'initial_weight', 'is_ffpe', 'lab_proc_her2_neu_immunohistochemistry_receptor_status', 'lab_procedure_her2_neu_in_situ_hybrid_outcome_type', 'lost_follow_up', 'lymph_node_examined_count', 'margin_status', 'menopause_status', 'metastatic_breast_carcinm_ps_fndng_prgstrn_rcptr_thr_msr_scl_txt', 'metastatic_breast_carcinom_lb_prc_hr2_n_mmnhstchmstry_rcptr_stts', 'metastatic_breast_carcinoma_erbb2_immunohistochemistry_levl_rslt', 'metastatic_breast_carcinoma_estrogen_receptor_detection_mthd_txt', 'metastatic_breast_carcinoma_estrogen_receptor_status', 'metastatic_breast_carcinoma_estrogen_receptr_lvl_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_her2_erbb_method_calculatin_mthd_txt', 'metastatic_breast_carcinoma_her2_erbb_pos_findng_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_her2_neu_chromosone_17_signal_rat_vl', 'metastatic_breast_carcinoma_immunhstchmstry_r_pstv_fndng_scl_typ', 'metastatic_breast_carcinoma_immunohistochemistry_er_pos_cell_scr', 'metastatic_breast_carcinoma_immunohistochemistry_pr_pos_cell_scr', 'metastatic_breast_carcinoma_lab_proc_hr2_n_n_st_hybrdztn_tcm_typ', 'metastatic_breast_carcinoma_pos_finding_hr2_rbb2_thr_msr_scl_txt', 'metastatic_breast_carcinoma_progestern_rcptr_lvl_cll_prcnt_ctgry', 'metastatic_breast_carcinoma_progesterone_receptor_dtctn_mthd_txt', 'metastatic_breast_carcinoma_progesterone_receptor_status', 'metastatic_site_at_diagnosis', 'metastatic_site_at_diagnosis_other', 'methylation_Clusters_nature2012', 'miRNA_Clusters_nature2012', 'mtsttc_brst_crcnm_flrscnc_n_st_hybrdztn_dgnstc_prc_cntrmr_17_sgn', 'mtsttc_brst_crcnm_hr2_rbb_ps_fndng_flrscnc_n_st_hybrdztn_clcltn', 'mtsttc_brst_crcnm_mmnhstchmstry_prgstrn_rcptr_pstv_fndng_scl_typ', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'number_of_lymphnodes_positive_by_he', 'number_of_lymphnodes_positive_by_ihc', 'oct_embedded', 'other_dx', 'pathologic_M', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_id', 'person_neoplasm_cancer_status', 'pgr_detection_method_text', 'pos_finding_her2_erbb2_other_measurement_scale_text', 'pos_finding_metastatic_brst_crcnm_strgn_rcptr_thr_msrmnt_scl_txt', 'pos_finding_progesterone_receptor_other_measurement_scale_text', 'positive_finding_estrogen_receptor_other_measurement_scale_text', 'postoperative_rx_tx', 'primary_lymph_node_presentation_assessment', 'progesterone_receptor_level_cell_percent_category', 'project_code', 'radiation_therapy', 'sample_type', 'sample_type_id', 'surgical_procedure_purpose_other_text', 'system_version', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'vial_number', 'vital_status', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_BRCA_RPPA_RBN', '_GENOMIC_ID_TCGA_BRCA_mutation', '_GENOMIC_ID_TCGA_BRCA_PDMRNAseq', '_GENOMIC_ID_TCGA_BRCA_hMethyl450', '_GENOMIC_ID_TCGA_BRCA_RPPA', '_GENOMIC_ID_TCGA_BRCA_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_BRCA_mutation_curated_wustl_gene', '_GENOMIC_ID_TCGA_BRCA_hMethyl27', '_GENOMIC_ID_TCGA_BRCA_PDMarrayCNV', '_GENOMIC_ID_TCGA_BRCA_miRNA_HiSeq', '_GENOMIC_ID_TCGA_BRCA_mutation_wustl_gene', '_GENOMIC_ID_TCGA_BRCA_miRNA_GA', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2_percentile', '_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_GA_gene', '_GENOMIC_ID_TCGA_BRCA_gistic2thd', '_GENOMIC_ID_data/public/TCGA/BRCA/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_BRCA_G4502A_07_3', '_GENOMIC_ID_TCGA_BRCA_exp_HiSeqV2', '_GENOMIC_ID_TCGA_BRCA_gistic2', '_GENOMIC_ID_TCGA_BRCA_PDMarray']\\n\""},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["f'''\n","Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n","candidate_age_cols = [col_name1, col_name2, ...]\n","candidate_gender_cols = [col_name1, col_name2, ...]\n","If no columns match a criterion, please provide an empty list.\n","\n","Column names:\n","{clinical_data_cols}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661858700Z","start_time":"2023-12-31T21:19:13.656749487Z"},"id":"X11W7KfziaD2","outputId":"ef120e77-44dd-443b-f86b-e464d962574c"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["candidate_age_cols = ['Age_at_Initial_Pathologic_Diagnosis_nature2012', 'age_at_initial_pathologic_diagnosis',\n","                      'days_to_birth', 'year_of_initial_pathologic_diagnosis']\n","candidate_gender_cols = ['Gender_nature2012', 'gender']"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661938833Z","start_time":"2023-12-31T21:19:13.656901848Z"},"id":"SicVNjZ4iaD2"}},{"cell_type":"markdown","source":["Choose a single column from the candidate columns that record age and gender information respectively.\n","If no column meets the requirement, keep 'age_col' or 'gender_col' to None"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:57:44.207565572Z","start_time":"2023-12-31T03:57:44.202177544Z"},"id":"APRmqURGiaD2"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"{'Age_at_Initial_Pathologic_Diagnosis_nature2012': [nan, nan, nan, nan, nan],\n 'age_at_initial_pathologic_diagnosis': [55.0, 50.0, 62.0, 52.0, 50.0],\n 'days_to_birth': [-20211.0, -18538.0, -22848.0, -19074.0, -18371.0],\n 'year_of_initial_pathologic_diagnosis': [2004.0,\n  2003.0,\n  2011.0,\n  2011.0,\n  2013.0]}"},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_age_cols])"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.662078139Z","start_time":"2023-12-31T21:19:13.657006017Z"},"id":"K6uQ_zrZiaD3","outputId":"1602b911-ac43-4f36-8ff6-5366b7bc573a"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["age_col = 'age_at_initial_pathologic_diagnosis'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.777159137Z","start_time":"2023-12-31T21:19:13.756210135Z"},"id":"ks4TtmqNiaD3"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"{'Gender_nature2012': [nan, nan, nan, nan, nan],\n 'gender': ['FEMALE', 'FEMALE', 'FEMALE', 'FEMALE', 'FEMALE']}"},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["preview_df(clinical_data[candidate_gender_cols])"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.313425023Z","start_time":"2023-12-31T21:19:14.305221062Z"},"id":"piIf5weZiaD3","outputId":"08a17b50-20d8-41bd-e1c7-fd8020193ceb"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["gender_col = 'gender'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.852502962Z","start_time":"2023-12-31T21:19:14.845153121Z"},"id":"tPNM-3cdiaD3"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col, gender_col)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.246065025Z","start_time":"2023-12-31T21:19:15.240279162Z"},"id":"Ed3TUm8_iaD3"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.593478652Z","start_time":"2023-12-31T21:19:15.583841656Z"},"id":"4BDE6fnKiaD3"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"                 Breast Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17  \\\nsampleID                                                                      \nTCGA-3C-AAAU-01              1  55.0     0.0   0.607308 -3.194126 -0.531035   \nTCGA-3C-AALI-01              1  50.0     0.0  -0.641192 -4.928226  0.095465   \nTCGA-3C-AALJ-01              1  62.0     0.0   1.082808 -4.623726 -0.531035   \nTCGA-3C-AALK-01              1  52.0     0.0   0.121608 -2.881526 -0.531035   \nTCGA-4H-AAAK-01              1  50.0     0.0   0.420208 -3.282726 -0.531035   \n\n                    RNF10     RNF11    RNF13   GTF2IP1  ...     TULP2  \\\nsampleID                                                ...             \nTCGA-3C-AAAU-01 -0.145872  0.237422 -0.29921 -0.142694  ... -0.748878   \nTCGA-3C-AALI-01  0.098128 -0.541978 -0.32291 -0.044694  ...  0.312922   \nTCGA-3C-AALJ-01  0.484028 -0.183678 -0.91901  0.261106  ... -0.748878   \nTCGA-3C-AALK-01  0.179128  0.039222 -0.45491  0.180306  ...  0.415422   \nTCGA-4H-AAAK-01 -0.020972 -0.117978 -0.55781 -0.173794  ...  0.685222   \n\n                    NPY5R     GNGT2    GNGT1     TULP3      PTRF     BCL6B  \\\nsampleID                                                                     \nTCGA-3C-AAAU-01  0.565583 -0.767233 -1.28139 -0.271377 -0.492286  0.360373   \nTCGA-3C-AALI-01  0.079683  0.475267  1.99661  0.161423  0.273714  0.644673   \nTCGA-3C-AALJ-01 -0.656117 -0.216733  0.21081  0.101023  0.660514  1.295073   \nTCGA-3C-AALK-01 -0.178417 -0.211233 -1.28139  0.108023  1.132814  0.496773   \nTCGA-4H-AAAK-01 -0.698717 -0.484233 -1.28139 -0.223577  1.222714  0.225573   \n\n                    GSTK1      SELP      SELS  \nsampleID                                       \nTCGA-3C-AAAU-01  1.067905  0.076267 -0.392212  \nTCGA-3C-AALI-01 -0.428695  0.068667 -0.043812  \nTCGA-3C-AALJ-01  0.915105  0.168567  0.047788  \nTCGA-3C-AALK-01  0.240105  3.099767  0.112888  \nTCGA-4H-AAAK-01 -0.301995  0.200067 -0.080212  \n\n[5 rows x 20533 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Breast Cancer</th>\n      <th>Age</th>\n      <th>Gender</th>\n      <th>ARHGEF10L</th>\n      <th>HIF3A</th>\n      <th>RNF17</th>\n      <th>RNF10</th>\n      <th>RNF11</th>\n      <th>RNF13</th>\n      <th>GTF2IP1</th>\n      <th>...</th>\n      <th>TULP2</th>\n      <th>NPY5R</th>\n      <th>GNGT2</th>\n      <th>GNGT1</th>\n      <th>TULP3</th>\n      <th>PTRF</th>\n      <th>BCL6B</th>\n      <th>GSTK1</th>\n      <th>SELP</th>\n      <th>SELS</th>\n    </tr>\n    <tr>\n      <th>sampleID</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>TCGA-3C-AAAU-01</th>\n      <td>1</td>\n      <td>55.0</td>\n      <td>0.0</td>\n      <td>0.607308</td>\n      <td>-3.194126</td>\n      <td>-0.531035</td>\n      <td>-0.145872</td>\n      <td>0.237422</td>\n      <td>-0.29921</td>\n      <td>-0.142694</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>0.565583</td>\n      <td>-0.767233</td>\n      <td>-1.28139</td>\n      <td>-0.271377</td>\n      <td>-0.492286</td>\n      <td>0.360373</td>\n      <td>1.067905</td>\n      <td>0.076267</td>\n      <td>-0.392212</td>\n    </tr>\n    <tr>\n      <th>TCGA-3C-AALI-01</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>-0.641192</td>\n      <td>-4.928226</td>\n      <td>0.095465</td>\n      <td>0.098128</td>\n      <td>-0.541978</td>\n      <td>-0.32291</td>\n      <td>-0.044694</td>\n      <td>...</td>\n      <td>0.312922</td>\n      <td>0.079683</td>\n      <td>0.475267</td>\n      <td>1.99661</td>\n      <td>0.161423</td>\n      <td>0.273714</td>\n      <td>0.644673</td>\n      <td>-0.428695</td>\n      <td>0.068667</td>\n      <td>-0.043812</td>\n    </tr>\n    <tr>\n      <th>TCGA-3C-AALJ-01</th>\n      <td>1</td>\n      <td>62.0</td>\n      <td>0.0</td>\n      <td>1.082808</td>\n      <td>-4.623726</td>\n      <td>-0.531035</td>\n      <td>0.484028</td>\n      <td>-0.183678</td>\n      <td>-0.91901</td>\n      <td>0.261106</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>-0.656117</td>\n      <td>-0.216733</td>\n      <td>0.21081</td>\n      <td>0.101023</td>\n      <td>0.660514</td>\n      <td>1.295073</td>\n      <td>0.915105</td>\n      <td>0.168567</td>\n      <td>0.047788</td>\n    </tr>\n    <tr>\n      <th>TCGA-3C-AALK-01</th>\n      <td>1</td>\n      <td>52.0</td>\n      <td>0.0</td>\n      <td>0.121608</td>\n      <td>-2.881526</td>\n      <td>-0.531035</td>\n      <td>0.179128</td>\n      <td>0.039222</td>\n      <td>-0.45491</td>\n      <td>0.180306</td>\n      <td>...</td>\n      <td>0.415422</td>\n      <td>-0.178417</td>\n      <td>-0.211233</td>\n      <td>-1.28139</td>\n      <td>0.108023</td>\n      <td>1.132814</td>\n      <td>0.496773</td>\n      <td>0.240105</td>\n      <td>3.099767</td>\n      <td>0.112888</td>\n    </tr>\n    <tr>\n      <th>TCGA-4H-AAAK-01</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>0.420208</td>\n      <td>-3.282726</td>\n      <td>-0.531035</td>\n      <td>-0.020972</td>\n      <td>-0.117978</td>\n      <td>-0.55781</td>\n      <td>-0.173794</td>\n      <td>...</td>\n      <td>0.685222</td>\n      <td>-0.698717</td>\n      <td>-0.484233</td>\n      <td>-1.28139</td>\n      <td>-0.223577</td>\n      <td>1.222714</td>\n      <td>0.225573</td>\n      <td>-0.301995</td>\n      <td>0.200067</td>\n      <td>-0.080212</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 20533 columns</p>\n</div>"},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n","merged_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.074598194Z","start_time":"2023-12-31T21:19:15.973600157Z"},"id":"Kao1VPxfiaD3","outputId":"a072e535-52f6-4ebf-8ba6-c2f59accfea7"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["The merged dataset contains 1215 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.341121723Z","start_time":"2023-12-31T21:19:16.330544222Z"},"id":"x7E8M0H-iaD3","outputId":"c14ee046-590a-4ea8-fcbb-1fcadca55632"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["For the feature 'Breast Cancer', the least common label is '0' with 113 occurrences. This represents 9.30% of the dataset.\n","The distribution of the feature 'Breast Cancer' in this dataset is fine.\n","\n","Quartiles for 'Age':\n","  25%: 48.0\n","  50% (Median): 58.0\n","  75%: 67.0\n","Min: 26.0\n","Max: 90.0\n","The distribution of the feature 'Age' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '1.0' with 13 occurrences. This represents 1.07% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"data":{"text/plain":"False"},"execution_count":18,"metadata":{},"output_type":"execute_result"}],"source":["is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:17.222505900Z","start_time":"2023-12-31T21:19:17.218781412Z"},"id":"YvBbzAe6iaD3","outputId":"15cf8038-a874-482d-af3d-d1204569e905"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.401536463Z","start_time":"2023-12-31T21:19:17.980683628Z"},"id":"Qzem-9-CiaD4"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.405288661Z","start_time":"2023-12-31T21:19:35.402799553Z"},"id":"6LuV7GjKiaD4"}},{"cell_type":"markdown","source":["## 2.2. The GEO dataset"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-28T21:28:38.028916303Z","start_time":"2023-12-28T21:28:38.016245426Z"},"id":"Kk0EYJ4yiaD4"}},{"cell_type":"markdown","source":["In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:25:23.253882615Z","start_time":"2023-12-31T03:25:23.244062710Z"},"id":"-RCIZQ6jiaD4"}},{"cell_type":"code","execution_count":5,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['GSE80060', 'GSE43553', 'GSE103500']"]},"metadata":{},"execution_count":5}],"source":["dataset = 'GEO'\n","trait_subdir = \"Autoinflammatory-Disorders\"\n","\n","trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n","os.listdir(trait_path)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.411177575Z","start_time":"2023-12-31T21:19:35.405710636Z"},"id":"zLCP2SDmiaD4","outputId":"fe593ad4-5134-410b-f0b2-e8e32b11e452","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459441002,"user_tz":-480,"elapsed":2642,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Repeat the below steps for all the accession numbers"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:39:42.634870142Z","start_time":"2023-12-31T03:39:42.534093295Z"},"id":"VHR6J86PiaD4"}},{"cell_type":"code","execution_count":6,"outputs":[{"output_type":"execute_result","data":{"text/plain":["('/content/drive/MyDrive/AI4Science_Public/data/GEO/Autoinflammatory-Disorders/GSE103500/GSE103500_family.soft.gz',\n"," '/content/drive/MyDrive/AI4Science_Public/data/GEO/Autoinflammatory-Disorders/GSE103500/GSE103500_series_matrix.txt.gz')"]},"metadata":{},"execution_count":6}],"source":["cohort = accession_num = \"GSE103500\"\n","cohort_dir = os.path.join(trait_path, accession_num)\n","soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n","soft_file, matrix_file"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.413836016Z","start_time":"2023-12-31T21:19:35.410443976Z"},"id":"-NttBC9AiaD4","outputId":"83884606-97d8-428c-9e4f-0c60104f30bf","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459441002,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["### Inital filtering and clinical data preprocessing"],"metadata":{"collapsed":false,"id":"OXII6n-eiaD4"}},{"cell_type":"code","execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["!Series_title\t\"Healthy adult and sJIA blood stimulation with innate immune ligands\"\n","!Series_summary\t\"The goal of this study was to characterize altered inducible immune networks in Systemic onset juvenile idiopathic arthritis (sJIA), an IL-1-driven autoinflammatory disease of unknown etiology. To this end, we developed a high-throughput assay that quantifies the transcriptional and protein-level responses of blood leukocytes to innate stimuli. Herein, we report transcriptional data from healthy adult blood stimulated with 16 different conditions, including TLR ligands, cytosolic receptor ligands and inflammatory cytokines. We further report blood transcriptional profiles from sJIA patients with various disease activity and treatment statuses, both ex vivo (baseline) and after in vitro stimulation with a subset of innate stimuli including heat-killed bacterial pathogens.\"\n","!Series_overall_design\t\"291 total samples, 31 donors and 17 stimuli (excluding media controls). All stimuli were not used with all donors. There is one technical replicate for donors BD1277, BD1350, BD1635 for all stimuli - the replicate was run in a later batch (larger barcode number).\"\n","!Series_overall_design\t\"\"\n","!Series_overall_design\t\"Stimuli used were agonists for surface Toll-like receptors (CpG-C, R837,  R848, TLR8L, Flagellin, LPS), a protein kinase C activator (PMA/I: phorbol myristate acetate and ionomycin), cytokines (IFNalpha, IFNgamma, IL-17A/F,  IL-18, IL-1beta, TNFalpha), gram+ derived ligands (PGN, LTA), and  heat-killed Staph. aureus (HKSA) as well as heat-killed Staph enteric (HKSE).\"\n","!Series_overall_design\t\"\"\n","!Series_overall_design\t\"Ligands were diluted in medium (RPMI 1640 with GlutaMax, Thermo Fisher  Scientific, Grand Island, NY, USA) and aliquoted into 2 mL polypropylene  96-well plates (Greiner Bio-One, Monroe, NC, USA). Blood was drawn in the  morning into the BD Vacutainers containing ACD anticoagulant (BD, Franklin  Lakes, NJ, USA). Time between blood draw and culture setup was less than  2h. Blood and appropriately diluted ligands were mixed in 1:1 ratio in  total volume of 1 mL/well and incubated for 6h at 37oC/5% CO2. After the culture, blood was mixed and  100 uL aliquot set aside for activation marker phenotyping and  intracellular cytokine staining (ICS) for IL-1b. The plate was then  centrifuged at 12,000 x g/15  and 400 mL of plasma aliquoted and stored at  -80oC until the Luminex analysis. The pellet was the  lysed with Tempus  solution (Tempus blood RNA tubes, Thermo Fisher Scientific) in 2:1 ratio  and stored at -20oC until the RNA extraction.\"\n","!Series_overall_design\t\"\"\n","!Series_overall_design\t\"Please note that data were analyzed using either V1 manifest (pre-Oct 2010) or the V2 manifest (post-Oct 2010) as indicated in the sample title.\"\n","!Series_overall_design\t\"\"\n","!Series_overall_design\t\"Sample Group abbreviation definitions:\"\n","!Series_overall_design\t\"sIU - SoJIA, Inactive, Untreated\"\n","!Series_overall_design\t\"sAU - SoJIA, Active, Untreated\"\n","!Series_overall_design\t\"sIU - SoJIA, Inactive, Anakinra\"\n"]}],"source":["background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n","clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n","\n","background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n","print(background_info)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.492119842Z","start_time":"2023-12-31T21:19:35.413383102Z"},"id":"zLBtCC4eiaD4","outputId":"e9955b06-f42c-4711-cd72-6b2294ed5ca4","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459444880,"user_tz":-480,"elapsed":1920,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":8,"outputs":[{"output_type":"execute_result","data":{"text/plain":["         !Sample_geo_accession                       GSM2124286  \\\n","0  !Sample_characteristics_ch1                 donor id: BD1350   \n","1  !Sample_characteristics_ch1    subject status: healthy adult   \n","2  !Sample_characteristics_ch1  disease_activity: healthy adult   \n","3  !Sample_characteristics_ch1                  treatment: none   \n","4  !Sample_characteristics_ch1     patient_group: healthy adult   \n","\n","                        GSM2124287                       GSM2124288  \\\n","0                 donor id: BD1635                 donor id: BD1635   \n","1    subject status: healthy adult    subject status: healthy adult   \n","2  disease_activity: healthy adult  disease_activity: healthy adult   \n","3                  treatment: none                  treatment: none   \n","4     patient_group: healthy adult     patient_group: healthy adult   \n","\n","                        GSM2124289                       GSM2124290  \\\n","0                 donor id: BD1350                 donor id: BD1277   \n","1    subject status: healthy adult    subject status: healthy adult   \n","2  disease_activity: healthy adult  disease_activity: healthy adult   \n","3                  treatment: none                  treatment: none   \n","4     patient_group: healthy adult     patient_group: healthy adult   \n","\n","                        GSM2124291                       GSM2124292  \\\n","0                 donor id: BD1350                 donor id: BD1350   \n","1    subject status: healthy adult    subject status: healthy adult   \n","2  disease_activity: healthy adult  disease_activity: healthy adult   \n","3                  treatment: none                  treatment: none   \n","4     patient_group: healthy adult     patient_group: healthy adult   \n","\n","                        GSM2124293                       GSM2124294  ...  \\\n","0                 donor id: BD1635                 donor id: BD1635  ...   \n","1    subject status: healthy adult    subject status: healthy adult  ...   \n","2  disease_activity: healthy adult  disease_activity: healthy adult  ...   \n","3                  treatment: none                  treatment: none  ...   \n","4     patient_group: healthy adult     patient_group: healthy adult  ...   \n","\n","                                          GSM2124577  \\\n","0                                  donor id: SYS-206   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124578  \\\n","0                                  donor id: SYS-188   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124579  \\\n","0                                  donor id: SYS-188   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124580  \\\n","0                                  donor id: SYS-188   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124581  \\\n","0                                  donor id: SYS-188   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124582  \\\n","0                                  donor id: SYS-130   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124583  \\\n","0                                  donor id: SYS-206   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124584  \\\n","0                                  donor id: SYS-206   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124585  \\\n","0                                  donor id: SYS-130   \n","1  subject status: Systemic onset juvenile idiopa...   \n","2                         disease_activity: inactive   \n","3                                treatment: Anakinra   \n","4                                 patient_group: sIA   \n","\n","                                          GSM2124586  \n","0                                  donor id: SYS-206  \n","1  subject status: Systemic onset juvenile idiopa...  \n","2                         disease_activity: inactive  \n","3                                treatment: Anakinra  \n","4                                 patient_group: sIA  \n","\n","[5 rows x 292 columns]"],"text/html":["\n","  <div id=\"df-ec1043a4-7c83-4e79-9f9b-f74c12298163\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>!Sample_geo_accession</th>\n","      <th>GSM2124286</th>\n","      <th>GSM2124287</th>\n","      <th>GSM2124288</th>\n","      <th>GSM2124289</th>\n","      <th>GSM2124290</th>\n","      <th>GSM2124291</th>\n","      <th>GSM2124292</th>\n","      <th>GSM2124293</th>\n","      <th>GSM2124294</th>\n","      <th>...</th>\n","      <th>GSM2124577</th>\n","      <th>GSM2124578</th>\n","      <th>GSM2124579</th>\n","      <th>GSM2124580</th>\n","      <th>GSM2124581</th>\n","      <th>GSM2124582</th>\n","      <th>GSM2124583</th>\n","      <th>GSM2124584</th>\n","      <th>GSM2124585</th>\n","      <th>GSM2124586</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>donor id: BD1350</td>\n","      <td>donor id: BD1635</td>\n","      <td>donor id: BD1635</td>\n","      <td>donor id: BD1350</td>\n","      <td>donor id: BD1277</td>\n","      <td>donor id: BD1350</td>\n","      <td>donor id: BD1350</td>\n","      <td>donor id: BD1635</td>\n","      <td>donor id: BD1635</td>\n","      <td>...</td>\n","      <td>donor id: SYS-206</td>\n","      <td>donor id: SYS-188</td>\n","      <td>donor id: SYS-188</td>\n","      <td>donor id: SYS-188</td>\n","      <td>donor id: SYS-188</td>\n","      <td>donor id: SYS-130</td>\n","      <td>donor id: SYS-206</td>\n","      <td>donor id: SYS-206</td>\n","      <td>donor id: SYS-130</td>\n","      <td>donor id: SYS-206</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>subject status: healthy adult</td>\n","      <td>...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","      <td>subject status: Systemic onset juvenile idiopa...</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>disease_activity: healthy adult</td>\n","      <td>...</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","      <td>disease_activity: inactive</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>treatment: none</td>\n","      <td>...</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","      <td>treatment: Anakinra</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>patient_group: healthy adult</td>\n","      <td>...</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","      <td>patient_group: sIA</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 292 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ec1043a4-7c83-4e79-9f9b-f74c12298163')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-ec1043a4-7c83-4e79-9f9b-f74c12298163 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-ec1043a4-7c83-4e79-9f9b-f74c12298163');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-b116b576-0b68-47cb-840c-b86d7e597002\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b116b576-0b68-47cb-840c-b86d7e597002')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-b116b576-0b68-47cb-840c-b86d7e597002 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":8}],"source":["clinical_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.507840422Z","start_time":"2023-12-31T21:19:35.451789107Z"},"id":"YwC2YMZ_iaD5","outputId":"790b0033-d324-44a9-a49a-01eefe1aeb45","colab":{"base_uri":"https://localhost:8080/","height":360},"executionInfo":{"status":"ok","timestamp":1704459446314,"user_tz":-480,"elapsed":5,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":9,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{0: ['donor id: BD1350',\n","  'donor id: BD1635',\n","  'donor id: BD1277',\n","  'donor id: SYS-109',\n","  'donor id: SYS-60',\n","  'donor id: SYS-180',\n","  'donor id: SYS-90',\n","  'donor id: SYS-110',\n","  'donor id: SYS-13',\n","  'donor id: SYS-53',\n","  'donor id: BD3038',\n","  'donor id: BD3045',\n","  'donor id: BD1312',\n","  'donor id: BD1768',\n","  'donor id: BD1737',\n","  'donor id: BD3041',\n","  'donor id: BD1858',\n","  'donor id: BD1837',\n","  'donor id: BD1063',\n","  'donor id: BD1658',\n","  'donor id: SYS-200',\n","  'donor id: H-386',\n","  'donor id: SYS-201',\n","  'donor id: H-287',\n","  'donor id: H-393',\n","  'donor id: H-390',\n","  'donor id: H-389',\n","  'donor id: H-391',\n","  'donor id: SYS-206',\n","  'donor id: SYS-130',\n","  'donor id: SYS-188'],\n"," 1: ['subject status: healthy adult',\n","  'subject status: Systemic onset juvenile idiopathic arthritis (sJIA)',\n","  'subject status: healthy pediatric'],\n"," 2: ['disease_activity: healthy adult',\n","  'disease_activity: inactive',\n","  'disease_activity: active',\n","  'disease_activity: healthy pediatric'],\n"," 3: ['treatment: none', 'treatment: untreated', 'treatment: Anakinra'],\n"," 4: ['patient_group: healthy adult',\n","  'patient_group: sIU',\n","  'patient_group: sIA',\n","  'patient_group: sAU',\n","  'patient_group: healthy pediatric'],\n"," 5: ['stimulus: R837',\n","  'stimulus: CpG-C',\n","  'stimulus: LTA',\n","  'stimulus: LPS',\n","  'stimulus: HKSA',\n","  'stimulus: Flagellin',\n","  'stimulus: PGN',\n","  'stimulus: IFNalpha',\n","  'stimulus: IL-1beta',\n","  'stimulus: HKSE',\n","  'stimulus: Medium',\n","  'stimulus: R848',\n","  'stimulus: IL-18',\n","  'stimulus: TLR8L',\n","  'stimulus: TNFalpha',\n","  'stimulus: IL-17A/F',\n","  'stimulus: PMA/I',\n","  'stimulus: IFNgamma'],\n"," 6: ['timepoint: 6h'],\n"," 7: ['tissue: whole blood']}"]},"metadata":{},"execution_count":9}],"source":["clinical_data_unique = get_unique_values_by_row(clinical_data)\n","clinical_data_unique"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508015426Z","start_time":"2023-12-31T21:19:35.492746637Z"},"id":"qUEEMODXiaD5","outputId":"530794df-97b8-40b0-fa0a-f94bf6850ee9","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459448203,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978204446Z","start_time":"2023-12-31T03:58:04.922270095Z"},"id":"zueh_J2PiaD5"}},{"cell_type":"code","execution_count":10,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Autoinflammatory Disorders\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Autoinflammatory Disorders\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Healthy adult and sJIA blood stimulation with innate immune ligands\"\\n!Series_summary\\t\"The goal of this study was to characterize altered inducible immune networks in Systemic onset juvenile idiopathic arthritis (sJIA), an IL-1-driven autoinflammatory disease of unknown etiology. To this end, we developed a high-throughput assay that quantifies the transcriptional and protein-level responses of blood leukocytes to innate stimuli. Herein, we report transcriptional data from healthy adult blood stimulated with 16 different conditions, including TLR ligands, cytosolic receptor ligands and inflammatory cytokines. We further report blood transcriptional profiles from sJIA patients with various disease activity and treatment statuses, both ex vivo (baseline) and after in vitro stimulation with a subset of innate stimuli including heat-killed bacterial pathogens.\"\\n!Series_overall_design\\t\"291 total samples, 31 donors and 17 stimuli (excluding media controls). All stimuli were not used with all donors. There is one technical replicate for donors BD1277, BD1350, BD1635 for all stimuli - the replicate was run in a later batch (larger barcode number).\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Stimuli used were agonists for surface Toll-like receptors (CpG-C, R837,  R848, TLR8L, Flagellin, LPS), a protein kinase C activator (PMA/I: phorbol myristate acetate and ionomycin), cytokines (IFNalpha, IFNgamma, IL-17A/F,  IL-18, IL-1beta, TNFalpha), gram+ derived ligands (PGN, LTA), and  heat-killed Staph. aureus (HKSA) as well as heat-killed Staph enteric (HKSE).\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Ligands were diluted in medium (RPMI 1640 with GlutaMax, Thermo Fisher  Scientific, Grand Island, NY, USA) and aliquoted into 2 mL polypropylene  96-well plates (Greiner Bio-One, Monroe, NC, USA). Blood was drawn in the  morning into the BD Vacutainers containing ACD anticoagulant (BD, Franklin  Lakes, NJ, USA). Time between blood draw and culture setup was less than  2h. Blood and appropriately diluted ligands were mixed in 1:1 ratio in  total volume of 1 mL/well and incubated for 6h at 37oC/5% CO2. After the culture, blood was mixed and  100 uL aliquot set aside for activation marker phenotyping and  intracellular cytokine staining (ICS) for IL-1b. The plate was then  centrifuged at 12,000 x g/15  and 400 mL of plasma aliquoted and stored at  -80oC until the Luminex analysis. The pellet was the  lysed with Tempus  solution (Tempus blood RNA tubes, Thermo Fisher Scientific) in 2:1 ratio  and stored at -20oC until the RNA extraction.\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Please note that data were analyzed using either V1 manifest (pre-Oct 2010) or the V2 manifest (post-Oct 2010) as indicated in the sample title.\"\\n!Series_overall_design\\t\"\"\\n!Series_overall_design\\t\"Sample Group abbreviation definitions:\"\\n!Series_overall_design\\t\"sIU - SoJIA, Inactive, Untreated\"\\n!Series_overall_design\\t\"sAU - SoJIA, Active, Untreated\"\\n!Series_overall_design\\t\"sIU - SoJIA, Inactive, Anakinra\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'donor id: BD1350\\', \\'donor id: BD1635\\', \\'donor id: BD1277\\', \\'donor id: SYS-109\\', \\'donor id: SYS-60\\', \\'donor id: SYS-180\\', \\'donor id: SYS-90\\', \\'donor id: SYS-110\\', \\'donor id: SYS-13\\', \\'donor id: SYS-53\\', \\'donor id: BD3038\\', \\'donor id: BD3045\\', \\'donor id: BD1312\\', \\'donor id: BD1768\\', \\'donor id: BD1737\\', \\'donor id: BD3041\\', \\'donor id: BD1858\\', \\'donor id: BD1837\\', \\'donor id: BD1063\\', \\'donor id: BD1658\\', \\'donor id: SYS-200\\', \\'donor id: H-386\\', \\'donor id: SYS-201\\', \\'donor id: H-287\\', \\'donor id: H-393\\', \\'donor id: H-390\\', \\'donor id: H-389\\', \\'donor id: H-391\\', \\'donor id: SYS-206\\', \\'donor id: SYS-130\\', \\'donor id: SYS-188\\'], 1: [\\'subject status: healthy adult\\', \\'subject status: Systemic onset juvenile idiopathic arthritis (sJIA)\\', \\'subject status: healthy pediatric\\'], 2: [\\'disease_activity: healthy adult\\', \\'disease_activity: inactive\\', \\'disease_activity: active\\', \\'disease_activity: healthy pediatric\\'], 3: [\\'treatment: none\\', \\'treatment: untreated\\', \\'treatment: Anakinra\\'], 4: [\\'patient_group: healthy adult\\', \\'patient_group: sIU\\', \\'patient_group: sIA\\', \\'patient_group: sAU\\', \\'patient_group: healthy pediatric\\'], 5: [\\'stimulus: R837\\', \\'stimulus: CpG-C\\', \\'stimulus: LTA\\', \\'stimulus: LPS\\', \\'stimulus: HKSA\\', \\'stimulus: Flagellin\\', \\'stimulus: PGN\\', \\'stimulus: IFNalpha\\', \\'stimulus: IL-1beta\\', \\'stimulus: HKSE\\', \\'stimulus: Medium\\', \\'stimulus: R848\\', \\'stimulus: IL-18\\', \\'stimulus: TLR8L\\', \\'stimulus: TNFalpha\\', \\'stimulus: IL-17A/F\\', \\'stimulus: PMA/I\\', \\'stimulus: IFNgamma\\'], 6: [\\'timepoint: 6h\\'], 7: [\\'tissue: whole blood\\']}\\n'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":10}],"source":["f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n","1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n","2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n","   (1) Is there human data available for this trait?\n","   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n","   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n","   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n","\n","Background information about the dataset:\n","{background_info}\n","\n","Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n","{clinical_data_unique}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508232400Z","start_time":"2023-12-31T21:19:35.492903487Z"},"id":"avSkGIZQiaD_","outputId":"3d9cf239-2a8d-41c1-c111-78cd92d031e4","colab":{"base_uri":"https://localhost:8080/","height":178},"executionInfo":{"status":"ok","timestamp":1704459451849,"user_tz":-480,"elapsed":5,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n","Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978240504Z","start_time":"2023-12-31T03:58:04.922365324Z"},"id":"WYm2ZV9ZiaD_"}},{"cell_type":"code","execution_count":11,"outputs":[],"source":["is_gene_availabe = True\n","trait_row_id = 1\n","age_row_id = None\n","gender_row_id = None\n","\n","trait_type = 'binary'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508295660Z","start_time":"2023-12-31T21:19:35.493015671Z"},"id":"PT7NFhWRiaEA","executionInfo":{"status":"ok","timestamp":1704459453619,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":12,"outputs":[],"source":["is_available = is_gene_availabe and (trait_row_id is not None)\n","if not is_available:\n","    save_cohort_info(cohort, JSON_PATH, is_available)\n","    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508365153Z","start_time":"2023-12-31T21:19:35.493151450Z"},"id":"gA7w5UyliaEA","executionInfo":{"status":"ok","timestamp":1704459467200,"user_tz":-480,"elapsed":457,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":13,"outputs":[],"source":["# Verify and use the functions generated by GPT\n","\n","def convert_trait(value):\n","    # Assuming 'subject status' represents Autoinflammatory Disorders\n","    if value == 'subject status: Systemic onset juvenile idiopathic arthritis (sJIA)':\n","        return 1  # Presence of Autoinflammatory Disorder (binary)\n","    else:\n","        return 0  # Absence of Autoinflammatory Disorder (binary)\n","\n","def convert_age(value):\n","    # Assuming age is continuous, handling unknown values\n","    try:\n","        return float(value)\n","    except (ValueError, TypeError):\n","        return None\n","\n","def convert_gender(value):\n","    # Assuming gender is binary (Male/Female), handling unknown values\n","    if value in ['Male', 'Female']:\n","        return value\n","    else:\n","        return None\n","\n","\n","\n"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508426069Z","start_time":"2023-12-31T21:19:35.493220101Z"},"id":"tY1MZ2cMiaEA","executionInfo":{"status":"ok","timestamp":1704459469360,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":14,"outputs":[{"output_type":"execute_result","data":{"text/plain":["                            GSM2124286  GSM2124287  GSM2124288  GSM2124289  \\\n","Autoinflammatory Disorders           0           0           0           0   \n","\n","                            GSM2124290  GSM2124291  GSM2124292  GSM2124293  \\\n","Autoinflammatory Disorders           0           0           0           0   \n","\n","                            GSM2124294  GSM2124295  ...  GSM2124577  \\\n","Autoinflammatory Disorders           0           0  ...           1   \n","\n","                            GSM2124578  GSM2124579  GSM2124580  GSM2124581  \\\n","Autoinflammatory Disorders           1           1           1           1   \n","\n","                            GSM2124582  GSM2124583  GSM2124584  GSM2124585  \\\n","Autoinflammatory Disorders           1           1           1           1   \n","\n","                            GSM2124586  \n","Autoinflammatory Disorders           1  \n","\n","[1 rows x 291 columns]"],"text/html":["\n","  <div id=\"df-0990348b-b748-4229-8a87-fab207701ebf\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GSM2124286</th>\n","      <th>GSM2124287</th>\n","      <th>GSM2124288</th>\n","      <th>GSM2124289</th>\n","      <th>GSM2124290</th>\n","      <th>GSM2124291</th>\n","      <th>GSM2124292</th>\n","      <th>GSM2124293</th>\n","      <th>GSM2124294</th>\n","      <th>GSM2124295</th>\n","      <th>...</th>\n","      <th>GSM2124577</th>\n","      <th>GSM2124578</th>\n","      <th>GSM2124579</th>\n","      <th>GSM2124580</th>\n","      <th>GSM2124581</th>\n","      <th>GSM2124582</th>\n","      <th>GSM2124583</th>\n","      <th>GSM2124584</th>\n","      <th>GSM2124585</th>\n","      <th>GSM2124586</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>Autoinflammatory Disorders</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1 rows × 291 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0990348b-b748-4229-8a87-fab207701ebf')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-0990348b-b748-4229-8a87-fab207701ebf button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-0990348b-b748-4229-8a87-fab207701ebf');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":14}],"source":["selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row_id, age_row_id, gender_row_id,\n","                                                      convert_trait, convert_age, convert_gender)\n","selected_clinical_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508533013Z","start_time":"2023-12-31T21:19:35.493272992Z"},"id":"6jgwzvo_iaEA","outputId":"9b0edf74-2d8a-4927-f90b-51e75a85dbea","colab":{"base_uri":"https://localhost:8080/","height":147},"executionInfo":{"status":"ok","timestamp":1704459471863,"user_tz":-480,"elapsed":466,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["### Genetic data preprocessing and final filtering"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978664193Z","start_time":"2023-12-31T03:58:04.966117261Z"},"id":"Avmg2XvQiaEA"}},{"cell_type":"code","execution_count":15,"outputs":[{"output_type":"execute_result","data":{"text/plain":["             ID  GSM2124286  GSM2124287  GSM2124288  GSM2124289  GSM2124290  \\\n","0  ILMN_1343291  20224.7148  16791.0059  20814.0977  20293.5547  18419.4316   \n","1  ILMN_1343295   1770.4265   1840.0784   1797.9597   2269.9917   1255.4275   \n","2  ILMN_1651199    -19.5596      9.3942     -6.4999     15.6880     -1.7502   \n","3  ILMN_1651209     -0.8812     -0.8292     22.2600      4.0365      1.0642   \n","4  ILMN_1651210     -1.7164     11.3661     15.6395      3.3921     17.9052   \n","\n","   GSM2124291  GSM2124292  GSM2124293  GSM2124294  ...  GSM2124577  \\\n","0  18530.1953  18748.1328  19658.6328  17679.5078  ...  21212.0586   \n","1   1750.5592   1913.2695   1833.3894   2521.4897  ...   1676.6591   \n","2    -14.5042      3.5772     -8.1621     -9.4599  ...     -4.2894   \n","3     16.0589     -1.0364     15.4457     16.5745  ...     25.8863   \n","4     25.2179      1.0190     13.5060      0.6578  ...      3.5027   \n","\n","   GSM2124578  GSM2124579  GSM2124580  GSM2124581  GSM2124582  GSM2124583  \\\n","0  20771.7207  18621.5625  16749.5781  16301.7588  15032.8750  17372.9395   \n","1   1423.4424   1605.4531    686.2298   1454.2601    712.8145   1317.3906   \n","2     -1.0172     -5.9130     -1.2286     -0.3214     -1.4056      1.4598   \n","3     21.7513     21.5471      5.4887     10.3351      5.1082     19.3818   \n","4      1.8973     -3.9909     -8.1034     -6.7846     -6.3385      6.1620   \n","\n","   GSM2124584  GSM2124585  GSM2124586  \n","0  17197.1113  16901.9336  18280.0508  \n","1   1167.7744    624.5963    868.8993  \n","2     -2.4067     -1.9414     -2.2119  \n","3     10.9213     11.0451      8.6737  \n","4      0.2771     -2.1866     -0.1293  \n","\n","[5 rows x 292 columns]"],"text/html":["\n","  <div id=\"df-6981be32-96b0-4286-87a7-25f7ee080f08\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ID</th>\n","      <th>GSM2124286</th>\n","      <th>GSM2124287</th>\n","      <th>GSM2124288</th>\n","      <th>GSM2124289</th>\n","      <th>GSM2124290</th>\n","      <th>GSM2124291</th>\n","      <th>GSM2124292</th>\n","      <th>GSM2124293</th>\n","      <th>GSM2124294</th>\n","      <th>...</th>\n","      <th>GSM2124577</th>\n","      <th>GSM2124578</th>\n","      <th>GSM2124579</th>\n","      <th>GSM2124580</th>\n","      <th>GSM2124581</th>\n","      <th>GSM2124582</th>\n","      <th>GSM2124583</th>\n","      <th>GSM2124584</th>\n","      <th>GSM2124585</th>\n","      <th>GSM2124586</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>ILMN_1343291</td>\n","      <td>20224.7148</td>\n","      <td>16791.0059</td>\n","      <td>20814.0977</td>\n","      <td>20293.5547</td>\n","      <td>18419.4316</td>\n","      <td>18530.1953</td>\n","      <td>18748.1328</td>\n","      <td>19658.6328</td>\n","      <td>17679.5078</td>\n","      <td>...</td>\n","      <td>21212.0586</td>\n","      <td>20771.7207</td>\n","      <td>18621.5625</td>\n","      <td>16749.5781</td>\n","      <td>16301.7588</td>\n","      <td>15032.8750</td>\n","      <td>17372.9395</td>\n","      <td>17197.1113</td>\n","      <td>16901.9336</td>\n","      <td>18280.0508</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>ILMN_1343295</td>\n","      <td>1770.4265</td>\n","      <td>1840.0784</td>\n","      <td>1797.9597</td>\n","      <td>2269.9917</td>\n","      <td>1255.4275</td>\n","      <td>1750.5592</td>\n","      <td>1913.2695</td>\n","      <td>1833.3894</td>\n","      <td>2521.4897</td>\n","      <td>...</td>\n","      <td>1676.6591</td>\n","      <td>1423.4424</td>\n","      <td>1605.4531</td>\n","      <td>686.2298</td>\n","      <td>1454.2601</td>\n","      <td>712.8145</td>\n","      <td>1317.3906</td>\n","      <td>1167.7744</td>\n","      <td>624.5963</td>\n","      <td>868.8993</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>ILMN_1651199</td>\n","      <td>-19.5596</td>\n","      <td>9.3942</td>\n","      <td>-6.4999</td>\n","      <td>15.6880</td>\n","      <td>-1.7502</td>\n","      <td>-14.5042</td>\n","      <td>3.5772</td>\n","      <td>-8.1621</td>\n","      <td>-9.4599</td>\n","      <td>...</td>\n","      <td>-4.2894</td>\n","      <td>-1.0172</td>\n","      <td>-5.9130</td>\n","      <td>-1.2286</td>\n","      <td>-0.3214</td>\n","      <td>-1.4056</td>\n","      <td>1.4598</td>\n","      <td>-2.4067</td>\n","      <td>-1.9414</td>\n","      <td>-2.2119</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>ILMN_1651209</td>\n","      <td>-0.8812</td>\n","      <td>-0.8292</td>\n","      <td>22.2600</td>\n","      <td>4.0365</td>\n","      <td>1.0642</td>\n","      <td>16.0589</td>\n","      <td>-1.0364</td>\n","      <td>15.4457</td>\n","      <td>16.5745</td>\n","      <td>...</td>\n","      <td>25.8863</td>\n","      <td>21.7513</td>\n","      <td>21.5471</td>\n","      <td>5.4887</td>\n","      <td>10.3351</td>\n","      <td>5.1082</td>\n","      <td>19.3818</td>\n","      <td>10.9213</td>\n","      <td>11.0451</td>\n","      <td>8.6737</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>ILMN_1651210</td>\n","      <td>-1.7164</td>\n","      <td>11.3661</td>\n","      <td>15.6395</td>\n","      <td>3.3921</td>\n","      <td>17.9052</td>\n","      <td>25.2179</td>\n","      <td>1.0190</td>\n","      <td>13.5060</td>\n","      <td>0.6578</td>\n","      <td>...</td>\n","      <td>3.5027</td>\n","      <td>1.8973</td>\n","      <td>-3.9909</td>\n","      <td>-8.1034</td>\n","      <td>-6.7846</td>\n","      <td>-6.3385</td>\n","      <td>6.1620</td>\n","      <td>0.2771</td>\n","      <td>-2.1866</td>\n","      <td>-0.1293</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 292 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6981be32-96b0-4286-87a7-25f7ee080f08')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-6981be32-96b0-4286-87a7-25f7ee080f08 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-6981be32-96b0-4286-87a7-25f7ee080f08');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-d68dfadd-5e9c-407e-add1-23f8fcc375db\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d68dfadd-5e9c-407e-add1-23f8fcc375db')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-d68dfadd-5e9c-407e-add1-23f8fcc375db button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":15}],"source":["genetic_data = get_genetic_data(matrix_file)\n","genetic_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590261560Z","start_time":"2023-12-31T21:19:35.493320362Z"},"id":"VnaNkDORiaEA","outputId":"fc8c9388-a369-4c8e-a945-9bfedde0f51f","colab":{"base_uri":"https://localhost:8080/","height":256},"executionInfo":{"status":"ok","timestamp":1704459490752,"user_tz":-480,"elapsed":2540,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":16,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['ILMN_1343291',\n"," 'ILMN_1343295',\n"," 'ILMN_1651199',\n"," 'ILMN_1651209',\n"," 'ILMN_1651210',\n"," 'ILMN_1651221',\n"," 'ILMN_1651228',\n"," 'ILMN_1651229',\n"," 'ILMN_1651230',\n"," 'ILMN_1651232',\n"," 'ILMN_1651235',\n"," 'ILMN_1651236',\n"," 'ILMN_1651237',\n"," 'ILMN_1651238',\n"," 'ILMN_1651249',\n"," 'ILMN_1651253',\n"," 'ILMN_1651254',\n"," 'ILMN_1651259',\n"," 'ILMN_1651260',\n"," 'ILMN_1651262']"]},"metadata":{},"execution_count":16}],"source":["gene_row_ids = genetic_data['ID'][:20].tolist()\n","gene_row_ids"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590443507Z","start_time":"2023-12-31T21:19:35.536713559Z"},"id":"aCWv6oYZiaEA","outputId":"3f614e5a-04c2-4972-8dbf-af619c0254f5","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459492414,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:30:41.595335164Z","start_time":"2023-12-31T03:30:41.513232329Z"},"id":"_GDJH2wbiaEA"}},{"cell_type":"markdown","source":["Reference prompt:"],"metadata":{"collapsed":false,"id":"-oGa4Zz0iaEA"}},{"cell_type":"code","execution_count":17,"outputs":[{"output_type":"execute_result","data":{"text/plain":["\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['ILMN_1343291', 'ILMN_1343295', 'ILMN_1651199', 'ILMN_1651209', 'ILMN_1651210', 'ILMN_1651221', 'ILMN_1651228', 'ILMN_1651229', 'ILMN_1651230', 'ILMN_1651232', 'ILMN_1651235', 'ILMN_1651236', 'ILMN_1651237', 'ILMN_1651238', 'ILMN_1651249', 'ILMN_1651253', 'ILMN_1651254', 'ILMN_1651259', 'ILMN_1651260', 'ILMN_1651262']\\n\""],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":17}],"source":["f'''\n","Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n","requires_gene_mapping = (True or False)\n","\n","Row headers:\n","{gene_row_ids}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590590427Z","start_time":"2023-12-31T21:19:35.536780517Z"},"id":"icIqZ_XaiaEB","outputId":"88837551-d222-43d9-8507-a2febc0aee22","colab":{"base_uri":"https://localhost:8080/","height":125},"executionInfo":{"status":"ok","timestamp":1704459495363,"user_tz":-480,"elapsed":515,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["If not required, jump directly to the gene normalization step"],"metadata":{"collapsed":false,"id":"lqpef-AEiaEB"}},{"cell_type":"code","execution_count":18,"outputs":[],"source":["requires_gene_mapping = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590648558Z","start_time":"2023-12-31T21:19:35.536821375Z"},"id":"mjjOtDejiaEB","executionInfo":{"status":"ok","timestamp":1704459497156,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":19,"outputs":[],"source":["if requires_gene_mapping:\n","    gene_annotation = get_gene_annotation(soft_file)[0]\n","    gene_annotation_summary = preview_df(gene_annotation)\n","    gene_annotation_summary"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590695457Z","start_time":"2023-12-31T21:19:35.536910836Z"},"id":"ddUHkPQeiaEB","executionInfo":{"status":"ok","timestamp":1704459540538,"user_tz":-480,"elapsed":41476,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978931927Z","start_time":"2023-12-31T03:58:04.966328339Z"},"id":"iX79MmEKiaEB"}},{"cell_type":"code","execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {'ID': ['ILMN_1343048', 'ILMN_1343049', 'ILMN_1343050', 'ILMN_1343052', 'ILMN_1343059'], 'Species': [nan, nan, nan, nan, nan], 'Source': [nan, nan, nan, nan, nan], 'Search_Key': [nan, nan, nan, nan, nan], 'Transcript': [nan, nan, nan, nan, nan], 'ILMN_Gene': [nan, nan, nan, nan, nan], 'Source_Reference_ID': [nan, nan, nan, nan, nan], 'RefSeq_ID': [nan, nan, nan, nan, nan], 'Unigene_ID': [nan, nan, nan, nan, nan], 'Entrez_Gene_ID': [nan, nan, nan, nan, nan], 'GI': [nan, nan, nan, nan, nan], 'Accession': [nan, nan, nan, nan, nan], 'Symbol': ['phage_lambda_genome', 'phage_lambda_genome', 'phage_lambda_genome:low', 'phage_lambda_genome:low', 'thrB'], 'Protein_Product': [nan, nan, nan, nan, 'thrB'], 'Probe_Id': [nan, nan, nan, nan, nan], 'Array_Address_Id': [5090180.0, 6510136.0, 7560739.0, 1450438.0, 1240647.0], 'Probe_Type': [nan, nan, nan, nan, nan], 'Probe_Start': [nan, nan, nan, nan, nan], 'SEQUENCE': ['GAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAA', 'CCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAA', 'CCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGAT', 'TCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGC', 'CTTGTGCCTGAGCTGTCAAAAGTAGAGCACGTCGCCGAGATGAAGGGCGC'], 'Chromosome': [nan, nan, nan, nan, nan], 'Probe_Chr_Orientation': [nan, nan, nan, nan, nan], 'Probe_Coordinates': [nan, nan, nan, nan, nan], 'Cytoband': [nan, nan, nan, nan, nan], 'Definition': [nan, nan, nan, nan, nan], 'Ontology_Component': [nan, nan, nan, nan, nan], 'Ontology_Process': [nan, nan, nan, nan, nan], 'Ontology_Function': [nan, nan, nan, nan, nan], 'Synonyms': [nan, nan, nan, nan, nan], 'Obsolete_Probe_Id': [nan, nan, nan, nan, nan], 'GB_ACC': [nan, nan, nan, nan, nan]}\n","    \n"]}],"source":["if requires_gene_mapping:\n","    print(f'''\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {gene_annotation_summary}\n","    ''')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590742236Z","start_time":"2023-12-31T21:19:35.537050893Z"},"id":"IFcuQ4S9iaEB","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459545733,"user_tz":-480,"elapsed":416,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"003b7f79-3587-425c-a76f-364f7668bcf5"}},{"cell_type":"code","execution_count":21,"outputs":[],"source":["if requires_gene_mapping:\n","    probe_id_key = 'ID'\n","    gene_symb_key = 'Symbol'\n","    gene_mapping = get_gene_mapping(gene_annotation, probe_id_key, gene_symb_key)\n","    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590786881Z","start_time":"2023-12-31T21:19:35.537132859Z"},"id":"pCVgeE6DiaEB","executionInfo":{"status":"ok","timestamp":1704459600983,"user_tz":-480,"elapsed":3382,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":22,"outputs":[{"output_type":"stream","name":"stderr","text":["INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('ABCC13', 2), ('ABCC6P1', 2), ('ABCC6P2', 3), ('ADAM6', 3), ('AGAP11', 2), ('AK2P2', 2), ('ALOX12P\n","WARNING:biothings.client:106 input query terms found no hit:\t['1-Dec', '1-Mar', '10-Mar', '11-Mar', '2-Mar', '3-Mar', '4-Mar', '5-Mar', '6-Mar', '7-Mar', '7A5', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:7 input query terms found dup hits:\t[('ATP6AP1L', 2), ('ATXN8OS', 2), ('BAGE2', 2), ('BMP2KL', 2), ('BRD7P2', 2), ('BRD7P3', 2), ('BRI3P\n","WARNING:biothings.client:360 input query terms found no hit:\t['ARMC4', 'ARMET', 'ARNTL', 'ARNTL2', 'ARP11', 'ARPM1', 'ARPM2', 'ARPP-21', 'ARS2', 'ARSE', 'ARVP612\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:3 input query terms found dup hits:\t[('C2orf27A', 2), ('C2orf83', 2), ('C5orf60', 2)]\n","WARNING:biothings.client:810 input query terms found no hit:\t['C15orf34', 'C15orf37', 'C15orf38', 'C15orf41', 'C15orf42', 'C15orf43', 'C15orf44', 'C15orf49', 'C1\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('CATSPER2P1', 2), ('CCDC144NL', 2), ('CCNYL2', 2), ('CCNYL3', 2), ('CCT6P1', 2), ('CDR1', 2), ('CE\n","WARNING:biothings.client:134 input query terms found no hit:\t['C9orf82', 'C9orf84', 'C9orf86', 'C9orf89', 'C9orf9', 'C9orf90', 'C9orf91', 'C9orf93', 'C9orf95', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:5 input query terms found dup hits:\t[('CLEC4GP1', 2), ('COPG2IT1', 2), ('CXADRP2', 3), ('CXADRP3', 2), ('CYP4Z2P', 2)]\n","WARNING:biothings.client:118 input query terms found no hit:\t['CLECL1', 'CLK2P', 'CLLU1OS', 'CLRN1OS', 'CMAH', 'CMYA3', 'CNIH', 'CNO', 'CNTD2', 'COBRA1', 'COL28'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:12 input query terms found dup hits:\t[('DEFB117', 2), ('DEFB122', 2), ('DHRS4L1', 3), ('DIRC1', 2), ('DLEU2L', 2), ('DNM1P35', 2), ('DPRX\n","WARNING:biothings.client:132 input query terms found no hit:\t['DEFB109', 'DEFB109P1', 'DEFB130', 'DEFB131', 'DEFB137', 'DEFB32', 'DEFB4', 'DEM1', 'DEPDC6', 'DFNA\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:7 input query terms found dup hits:\t[('FAM197Y2', 3), ('FAM41AY2', 2), ('FAM74A3', 2), ('FAM90A2P', 2), ('FAM90A6P', 3), ('FCGR2C', 2), \n","WARNING:biothings.client:483 input query terms found no hit:\t['EVI1', 'EXOC3L', 'EXOD1', 'FABP5L2', 'FABP5L3', 'FABP5L7', 'FAIM3', 'FALZ', 'FAM100A', 'FAM100B', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:7 input query terms found dup hits:\t[('FMO9P', 2), ('FOLH1B', 2), ('GABPAP', 2), ('GAGE12B', 2), ('GGT3P', 2), ('GGT8P', 2), ('GLRA4', 2\n","WARNING:biothings.client:230 input query terms found no hit:\t['FLJ41200', 'FLJ41327', 'FLJ41352', 'FLJ41423', 'FLJ41481', 'FLJ41484', 'FLJ41562', 'FLJ41603', 'FL\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('GTF2H2B', 2), ('GTF2IP1', 2), ('HAR1A', 2), ('HBBP1', 2), ('HBD', 2), ('HCG4', 9), ('HERC2P2', 3)\n","WARNING:biothings.client:206 input query terms found no hit:\t['GPR97', 'GPR98', 'GRAMD2', 'GRAMD3', 'GRASP', 'GRINL1A', 'GRLF1', 'GRRP1', 'GS85', 'GSDM1', 'GSG2'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('KIR3DP1', 10), ('KIR3DX1', 5), ('KLKP1', 2), ('KRT126P', 2), ('KRT17P3', 2), ('KRT18P13', 2), ('K\n","WARNING:biothings.client:218 input query terms found no hit:\t['IHPK3', 'IIP45', 'IKBKAP', 'IL1F5', 'IL1F6', 'IL1F7', 'IL1F8', 'IL1F9', 'IL28A', 'IL28B', 'IL28RA'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:2 input query terms found dup hits:\t[('KRT8P15', 2), ('KRT8P9', 2)]\n","WARNING:biothings.client:597 input query terms found no hit:\t['KRTAP9-5', 'KRTHA5', 'KRTHB6', 'KTELC1', 'L3MBTL', 'LACE1', 'LAMB2L', 'LARGE', 'LARS', 'LASS1', 'L\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:956 input query terms found no hit:\t['LOC100129876', 'LOC100129878', 'LOC100129882', 'LOC100129887', 'LOC100129890', 'LOC100129901', 'LO\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:965 input query terms found no hit:\t['LOC100133017', 'LOC100133018', 'LOC100133019', 'LOC100133020', 'LOC100133021', 'LOC100133025', 'LO\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:925 input query terms found no hit:\t['LOC284648', 'LOC284672', 'LOC284688', 'LOC284701', 'LOC284749', 'LOC284751', 'LOC284757', 'LOC2848\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:948 input query terms found no hit:\t['LOC440456', 'LOC440459', 'LOC440461', 'LOC440487', 'LOC440491', 'LOC440498', 'LOC440503', 'LOC4405\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:966 input query terms found no hit:\t['LOC643272', 'LOC643273', 'LOC643274', 'LOC643275', 'LOC643278', 'LOC643283', 'LOC643284', 'LOC6432\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:961 input query terms found no hit:\t['LOC645463', 'LOC645464', 'LOC645465', 'LOC645466', 'LOC645476', 'LOC645478', 'LOC645487', 'LOC6454\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:994 input query terms found no hit:\t['LOC648249', 'LOC648251', 'LOC648253', 'LOC648256', 'LOC648257', 'LOC648262', 'LOC648264', 'LOC6482\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:990 input query terms found no hit:\t['LOC651900', 'LOC651905', 'LOC651907', 'LOC651914', 'LOC651916', 'LOC651919', 'LOC651921', 'LOC6519\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:944 input query terms found no hit:\t['LOC727844', 'LOC727846', 'LOC727847', 'LOC727848', 'LOC727849', 'LOC727851', 'LOC727855', 'LOC7278\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:7 input query terms found dup hits:\t[('LPAL2', 2), ('LRP5L', 2), ('LY6G6E', 8), ('LYPLA2P1', 6), ('MALAT1', 2), ('MEG8', 2), ('MEIS3P1',\n","WARNING:biothings.client:364 input query terms found no hit:\t['LOC730796', 'LOC730804', 'LOC730805', 'LOC730809', 'LOC730811', 'LOC730818', 'LOC730820', 'LOC7308\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('MIR320A', 2), ('MIR320B1', 2), ('MIR320B2', 2), ('MIR320C1', 2), ('MMD2', 2), ('MMP23A', 2), ('MO\n","WARNING:biothings.client:105 input query terms found no hit:\t['MGC45922', 'MGC46336', 'MGC4677', 'MGC48628', 'MGC48637', 'MGC50722', 'MGC51338', 'MGC5139', 'MGC5\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:10 input query terms found dup hits:\t[('MYH16', 2), ('NAPSB', 2), ('NAT8B', 3), ('NBPF22P', 2), ('NCF1B', 2), ('NME2P1', 2), ('NUDT9P1', \n","WARNING:biothings.client:122 input query terms found no hit:\t['MTL5', 'MTM', 'MTMR15', 'MTP18', 'MTSS1L', 'MUCDHL', 'MUDENG', 'MUM1', 'MUM1L1', 'MUPCDH', 'MURC',\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:10 input query terms found dup hits:\t[('OR10J3', 3), ('OR2A9P', 2), ('OR6W1P', 3), ('OR7E156P', 2), ('OR7E37P', 2), ('OR7E91P', 2), ('P2R\n","WARNING:biothings.client:64 input query terms found no hit:\t['OPRS1', 'OR2B3P', 'OR2W5', 'OR3A4', 'OR52A4', 'OR5AY1', 'OR5R1', 'OR8G2', 'ORAOV1', 'ORC1L', 'ORC2\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('PGM5P2', 2), ('PI4KAP1', 2), ('PI4KAP2', 2), ('PIPSL', 2), ('PKD1P1', 3), ('PLGLA', 2), ('PMCHL1'\n","WARNING:biothings.client:102 input query terms found no hit:\t['PH-4', 'PHACS', 'PHB', 'PHCA', 'PHF15', 'PHF16', 'PHF17', 'PIB5PA', 'PIK4CA', 'PIN1L', 'PION', 'PI\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:22 input query terms found dup hits:\t[('PSORS1C3', 8), ('PTTG3P', 2), ('PVT1', 2), ('PYY2', 2), ('RAET1K', 2), ('RAP1BL', 2), ('RBMY1A3P'\n","WARNING:biothings.client:93 input query terms found no hit:\t['PSPHL', 'PTCHD2', 'PTK9', 'PTPLA', 'PTPLAD1', 'PTPLAD2', 'PTPLB', 'PTPN20A', 'PTPN20B', 'PTPRV', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:13 input query terms found dup hits:\t[('RPPH1', 3), ('RPS10P3', 2), ('RPS26P11', 2), ('RRN3P2', 2), ('S100A7L2', 2), ('SCARNA17', 2), ('S\n","WARNING:biothings.client:100 input query terms found no hit:\t['RPRC1', 'RPS26L', 'RQCD1', 'RRP7B', 'RSHL1', 'RSHL3', 'RTCD1', 'RTDR1', 'RUNDC2A', 'RUNDC2B', 'RUN\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:63 input query terms found dup hits:\t[('SLC6A10P', 2), ('SLC7A5P1', 2), ('SNHG5', 2), ('SNHG6', 2), ('SNHG8', 2), ('SNHG9', 3), ('SNORA16\n","WARNING:biothings.client:62 input query terms found no hit:\t['SLC35E2', 'SLC9A10', 'SLC9A11', 'SLC9A3R1', 'SLC9A3R2', 'SLIC1', 'SLMO1', 'SLMO2', 'SMA3', 'SMA4',\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:12 input query terms found dup hits:\t[('ST20', 2), ('STAG3L1', 2), ('STAG3L2', 2), ('STAG3L3', 2), ('STAG3L4', 2), ('SUMO1P1', 2), ('SUMO\n","WARNING:biothings.client:118 input query terms found no hit:\t['ST5', 'ST7OT1', 'ST7OT2', 'ST7OT3', 'ST7OT4', 'STELLAR', 'STGC3', 'STRA13', 'STS-1', 'SUGT1L1', 'S\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('TNXA', 5), ('TOP1P1', 2), ('TP53TG1', 2), ('TRIM16L', 2), ('TRPC2', 2), ('TTTY11', 2), ('TTTY17C'\n","WARNING:biothings.client:105 input query terms found no hit:\t['TMEM211', 'TMEM22', 'TMEM27', 'TMEM34', 'TMEM35', 'TMEM4', 'TMEM48', 'TMEM49', 'TMEM5', 'TMEM55A',\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:4 input query terms found dup hits:\t[('VENTXP7', 2), ('WASH2P', 2), ('WASH3P', 3), ('XKRY', 2)]\n","WARNING:biothings.client:123 input query terms found no hit:\t['USH3A', 'USMG5', 'USP17', 'UTP11L', 'UTS2D', 'UTX', 'VARS', 'VCX-C', 'VIL2', 'VISA', 'VMD2L3', 'VN\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-334...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:3 input query terms found dup hits:\t[('ZNF658B', 2), ('ZNF781', 2), ('rab1c', 2)]\n","WARNING:biothings.client:30 input query terms found no hit:\t['ZNF509', 'ZNF533', 'ZNF542', 'ZNF642', 'ZNF643', 'ZNF645', 'ZNF650', 'ZNF663', 'ZNF673', 'ZNF720',\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"]}],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590837318Z","start_time":"2023-12-31T21:19:35.537209586Z"},"id":"aqJRhdbFiaEB","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459682112,"user_tz":-480,"elapsed":79400,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"e32013d3-5c4b-4dd3-ec63-603a958dd08d"}},{"cell_type":"code","execution_count":23,"outputs":[],"source":["merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n","# The preprocessing runs through, which means is_available should be True\n","is_available = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590902191Z","start_time":"2023-12-31T21:19:35.537285901Z"},"id":"istECoP1iaEB","executionInfo":{"status":"ok","timestamp":1704459686452,"user_tz":-480,"elapsed":410,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":24,"outputs":[{"output_type":"stream","name":"stdout","text":["The merged dataset contains 291 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.591095170Z","start_time":"2023-12-31T21:19:35.538708445Z"},"id":"tzrpD6LciaEB","outputId":"ee927911-f436-4e12-91b7-85e031ccd63b","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459689000,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":25,"outputs":[{"output_type":"stream","name":"stdout","text":["For the feature 'Autoinflammatory Disorders', the least common label is '1.0' with 93 occurrences. This represents 31.96% of the dataset.\n","The distribution of the feature 'Autoinflammatory Disorders' in this dataset is fine.\n","\n"]},{"output_type":"execute_result","data":{"text/plain":["False"]},"metadata":{},"execution_count":25}],"source":["is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610849182Z","start_time":"2023-12-31T21:19:35.541210258Z"},"id":"opVFXzd4iaEC","outputId":"aeee804a-52e1-4699-bcc4-e074b2b9a2ee","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704459690597,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":26,"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610923234Z","start_time":"2023-12-31T21:19:35.584848473Z"},"id":"wFORrTwciaEC","executionInfo":{"status":"ok","timestamp":1704459693473,"user_tz":-480,"elapsed":949,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":27,"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610970734Z","start_time":"2023-12-31T21:19:35.585042694Z"},"id":"EG3DaeSJiaEC","executionInfo":{"status":"ok","timestamp":1704459700767,"user_tz":-480,"elapsed":5441,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["### 3. Do regression & Cross Validation"],"metadata":{"id":"-MTPhRGxJV7I","ExecuteTime":{"end_time":"2023-12-31T03:58:22.271683755Z","start_time":"2023-12-31T03:58:22.246557674Z"}}},{"cell_type":"code","execution_count":28,"outputs":[],"source":["best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH, 'has_age')\n","best_cohort"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.459442087Z","start_time":"2023-12-31T21:19:44.416219648Z"},"id":"v-dUlz9_iaEC","executionInfo":{"status":"ok","timestamp":1704459733416,"user_tz":-480,"elapsed":533,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"   cohort_id  is_usable  is_available is_biased has_age has_gender  \\\n0       Xena       True          True     False    True       True   \n3  GSE248830       True          True     False    True       True   \n\n   sample_size note  \n0       1215.0       \n3         38.0       ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>cohort_id</th>\n      <th>is_usable</th>\n      <th>is_available</th>\n      <th>is_biased</th>\n      <th>has_age</th>\n      <th>has_gender</th>\n      <th>sample_size</th>\n      <th>note</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Xena</td>\n      <td>True</td>\n      <td>True</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>1215.0</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>GSE248830</td>\n      <td>True</td>\n      <td>True</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>38.0</td>\n      <td></td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"execution_count":45,"metadata":{},"output_type":"execute_result"}],"source":["ranked_df.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.808945960Z","start_time":"2023-12-31T21:19:44.799467072Z"},"id":"r15ErhKhiaEC","outputId":"cf26557f-7ea5-4980-d4f5-e77fe72b50da"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, best_cohort+'.csv'))"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.222719978Z","start_time":"2023-12-31T21:19:45.162873006Z"},"id":"2c15dYzIiaEC"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"   Breast Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17     RNF10  \\\n0              1  55.0     0.0   0.607308 -3.194126 -0.531035 -0.145872   \n1              1  50.0     0.0  -0.641192 -4.928226  0.095465  0.098128   \n2              1  62.0     0.0   1.082808 -4.623726 -0.531035  0.484028   \n3              1  52.0     0.0   0.121608 -2.881526 -0.531035  0.179128   \n4              1  50.0     0.0   0.420208 -3.282726 -0.531035 -0.020972   \n\n      RNF11    RNF13   GTF2IP1  ...     TULP2     NPY5R     GNGT2    GNGT1  \\\n0  0.237422 -0.29921 -0.142694  ... -0.748878  0.565583 -0.767233 -1.28139   \n1 -0.541978 -0.32291 -0.044694  ...  0.312922  0.079683  0.475267  1.99661   \n2 -0.183678 -0.91901  0.261106  ... -0.748878 -0.656117 -0.216733  0.21081   \n3  0.039222 -0.45491  0.180306  ...  0.415422 -0.178417 -0.211233 -1.28139   \n4 -0.117978 -0.55781 -0.173794  ...  0.685222 -0.698717 -0.484233 -1.28139   \n\n      TULP3      PTRF     BCL6B     GSTK1      SELP      SELS  \n0 -0.271377 -0.492286  0.360373  1.067905  0.076267 -0.392212  \n1  0.161423  0.273714  0.644673 -0.428695  0.068667 -0.043812  \n2  0.101023  0.660514  1.295073  0.915105  0.168567  0.047788  \n3  0.108023  1.132814  0.496773  0.240105  3.099767  0.112888  \n4 -0.223577  1.222714  0.225573 -0.301995  0.200067 -0.080212  \n\n[5 rows x 20533 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Breast Cancer</th>\n      <th>Age</th>\n      <th>Gender</th>\n      <th>ARHGEF10L</th>\n      <th>HIF3A</th>\n      <th>RNF17</th>\n      <th>RNF10</th>\n      <th>RNF11</th>\n      <th>RNF13</th>\n      <th>GTF2IP1</th>\n      <th>...</th>\n      <th>TULP2</th>\n      <th>NPY5R</th>\n      <th>GNGT2</th>\n      <th>GNGT1</th>\n      <th>TULP3</th>\n      <th>PTRF</th>\n      <th>BCL6B</th>\n      <th>GSTK1</th>\n      <th>SELP</th>\n      <th>SELS</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1</td>\n      <td>55.0</td>\n      <td>0.0</td>\n      <td>0.607308</td>\n      <td>-3.194126</td>\n      <td>-0.531035</td>\n      <td>-0.145872</td>\n      <td>0.237422</td>\n      <td>-0.29921</td>\n      <td>-0.142694</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>0.565583</td>\n      <td>-0.767233</td>\n      <td>-1.28139</td>\n      <td>-0.271377</td>\n      <td>-0.492286</td>\n      <td>0.360373</td>\n      <td>1.067905</td>\n      <td>0.076267</td>\n      <td>-0.392212</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>-0.641192</td>\n      <td>-4.928226</td>\n      <td>0.095465</td>\n      <td>0.098128</td>\n      <td>-0.541978</td>\n      <td>-0.32291</td>\n      <td>-0.044694</td>\n      <td>...</td>\n      <td>0.312922</td>\n      <td>0.079683</td>\n      <td>0.475267</td>\n      <td>1.99661</td>\n      <td>0.161423</td>\n      <td>0.273714</td>\n      <td>0.644673</td>\n      <td>-0.428695</td>\n      <td>0.068667</td>\n      <td>-0.043812</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>1</td>\n      <td>62.0</td>\n      <td>0.0</td>\n      <td>1.082808</td>\n      <td>-4.623726</td>\n      <td>-0.531035</td>\n      <td>0.484028</td>\n      <td>-0.183678</td>\n      <td>-0.91901</td>\n      <td>0.261106</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>-0.656117</td>\n      <td>-0.216733</td>\n      <td>0.21081</td>\n      <td>0.101023</td>\n      <td>0.660514</td>\n      <td>1.295073</td>\n      <td>0.915105</td>\n      <td>0.168567</td>\n      <td>0.047788</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>1</td>\n      <td>52.0</td>\n      <td>0.0</td>\n      <td>0.121608</td>\n      <td>-2.881526</td>\n      <td>-0.531035</td>\n      <td>0.179128</td>\n      <td>0.039222</td>\n      <td>-0.45491</td>\n      <td>0.180306</td>\n      <td>...</td>\n      <td>0.415422</td>\n      <td>-0.178417</td>\n      <td>-0.211233</td>\n      <td>-1.28139</td>\n      <td>0.108023</td>\n      <td>1.132814</td>\n      <td>0.496773</td>\n      <td>0.240105</td>\n      <td>3.099767</td>\n      <td>0.112888</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>0.420208</td>\n      <td>-3.282726</td>\n      <td>-0.531035</td>\n      <td>-0.020972</td>\n      <td>-0.117978</td>\n      <td>-0.55781</td>\n      <td>-0.173794</td>\n      <td>...</td>\n      <td>0.685222</td>\n      <td>-0.698717</td>\n      <td>-0.484233</td>\n      <td>-1.28139</td>\n      <td>-0.223577</td>\n      <td>1.222714</td>\n      <td>0.225573</td>\n      <td>-0.301995</td>\n      <td>0.200067</td>\n      <td>-0.080212</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 20533 columns</p>\n</div>"},"execution_count":47,"metadata":{},"output_type":"execute_result"}],"source":["merged_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.234945878Z","start_time":"2023-12-31T21:19:52.231604618Z"},"id":"LOAj-G92iaEC","outputId":"519f4ae4-7c39-4ba0-cbb4-f2d4f67dcc44"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["# If both age and gender features are available, select 'age' as the condition.\n","condition = 'Age'\n","# Remove the other condition to prevent interference.\n","merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.312540150Z","start_time":"2023-12-31T21:19:52.235772334Z"},"id":"vtzuVZpLiaEC"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["X = merged_data.drop(columns=[TRAIT, condition]).values\n","Y = merged_data[TRAIT].values\n","Z = merged_data[condition].values"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.323468266Z","start_time":"2023-12-31T21:19:52.290903989Z"},"id":"LJDde0WQiaEC"}},{"cell_type":"markdown","source":["Select the appropriate regression model depending on whether the dataset shows batch effect."],"metadata":{"collapsed":false,"id":"VKBCcPkGiaEC"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"True"},"execution_count":50,"metadata":{},"output_type":"execute_result"}],"source":["has_batch_effect = detect_batch_effect(X)\n","has_batch_effect"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.930840782Z","start_time":"2023-12-31T21:19:52.319092572Z"},"id":"0mp8bOqciaED","outputId":"2e542440-11b5-4751-9d72-460b69859c8f"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["if has_batch_effect:\n","    model_constructor = VariableSelection\n","    model_params = {}\n","else:\n","    model_constructor = Lasso\n","    model_params = {'alpha': 1.0, 'random_state': 42}"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.931029722Z","start_time":"2023-12-31T21:20:00.929710828Z"},"id":"IITCrXFGiaED"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stderr","output_type":"stream","text":["/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n"]},{"name":"stdout","output_type":"stream","text":["The cross-validation accuracy is 91.28% ± 1.50%\n"]}],"source":["cv_mean, cv_std = cross_validation(X, Y, Z, model_constructor, model_params, target_type=trait_type)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:20.709459824Z","start_time":"2023-12-31T21:20:00.931657949Z"},"id":"qWZ4p5LRiaED","outputId":"94d8262b-dfc7-44fc-8a36-b36a140bcd94"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stderr","output_type":"stream","text":["/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n"]}],"source":["# Train regression model on the whole dataset to identify significant genes\n","model = ResidualizationRegressor(model_constructor, model_params)\n","normalized_X, _ = normalize_data(X)\n","normalized_Z, _ = normalize_data(Z)\n","model.fit(normalized_X, Y, normalized_Z)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:24.980870159Z","start_time":"2023-12-31T21:20:20.516710807Z"},"id":"6PpgQE0kiaED","outputId":"0466e0eb-88f5-4e9a-8236-666d416f1b1d"}},{"cell_type":"markdown","source":["### 5. Discussion and report"],"metadata":{"id":"EjJrxbvb4nlj","ExecuteTime":{"end_time":"2023-10-14T04:50:59.072075075Z","start_time":"2023-10-14T04:50:38.739499998Z"}}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["Effect of the condition on the target variable:\n","Variable: Age\n","Coefficient: 0.0065\n","p-value: nan\n","\n","Found 76 significant genes associated with the trait 'Breast Cancer' conditional on the factor 'Age', with corrected p-value < 0.05:\n","    Variable  Coefficient  corrected_p_value\n","     COL10A1     0.136339       3.404090e-27\n","         BGN     0.095422       1.004817e-14\n","    PPAPDC1A     0.081296       9.785054e-13\n","       MMP11     0.080354       1.123019e-10\n","       WISP1     0.074888       9.613830e-10\n","     HSD17B6     0.051940       6.980811e-08\n","      CTHRC1     0.058310       2.210553e-07\n","       MFAP5    -0.054250       2.706783e-07\n","      CLEC5A     0.047846       8.877914e-07\n","   C20orf103     0.048957       8.877914e-07\n","          GC     0.027375       1.419502e-05\n","      OR51D1     0.022908       1.644207e-05\n","        COMP     0.041129       4.366228e-05\n","LOC100169752     0.019398       4.366228e-05\n","      LRRC15     0.052746       4.494130e-05\n","         FN1     0.063527       5.027562e-05\n","     C1QTNF6     0.049130       1.220626e-04\n","       INHBA     0.052621       1.384526e-04\n","       FNDC1     0.044080       1.724540e-04\n","        VCAN     0.062919       3.541404e-04\n","   LOC283867     0.039438       3.541404e-04\n","     SLC24A2     0.045359       3.929641e-04\n","        ECM2    -0.066303       4.895350e-04\n","       CILP2     0.040613       4.895350e-04\n","         DCN    -0.062282       5.405761e-04\n","        HEYL     0.039376       6.133304e-04\n","         FAP     0.062297       6.903458e-04\n","     OLFML2A    -0.037988       8.815285e-04\n","       MFAP2     0.047162       8.815285e-04\n","       CCRL1    -0.036802       8.815285e-04\n","       CD248    -0.046771       8.815285e-04\n","       P4HA3     0.044037       8.815285e-04\n","     C5orf13     0.037746       1.590045e-03\n","      COL5A2     0.064118       2.057973e-03\n","         F2R     0.041164       2.101197e-03\n","      OLFML1    -0.061570       3.573228e-03\n","       PRDM8    -0.040662       3.786028e-03\n","   POM121L12     0.028947       3.941936e-03\n","       LOXL1     0.044681       4.547038e-03\n","       SPARC     0.053330       5.391493e-03\n","      LEFTY2     0.027338       5.753501e-03\n","    SERPINE1     0.028865       5.753501e-03\n","       OR2L3    -0.016537       6.402547e-03\n","     CYP19A1     0.022029       6.651093e-03\n","         MAF    -0.038952       7.098913e-03\n","       GNG11    -0.068155       1.242228e-02\n","       CFHR5    -0.018470       1.366939e-02\n","      CLEC3B    -0.043039       1.366939e-02\n","        NRG3    -0.019375       1.366939e-02\n","      COL1A1     0.046342       1.427700e-02\n","      PCOLCE    -0.049663       1.836229e-02\n","     TMEM119     0.037238       1.922482e-02\n","       KCNC1    -0.024520       2.248216e-02\n","       WNT8A     0.014598       2.592754e-02\n","        RECK    -0.052887       2.781665e-02\n","        SNCB     0.018240       3.140935e-02\n","      KBTBD6    -0.029237       3.140935e-02\n","       SRPX2    -0.034794       3.140935e-02\n","     OLFML2B     0.036407       3.214194e-02\n","      PLXDC1    -0.032863       3.219694e-02\n","      ADAM12     0.040368       3.309987e-02\n","      GRIN2A    -0.021383       3.340752e-02\n","      COL5A1     0.050110       3.456174e-02\n","      USHBP1    -0.050144       3.618703e-02\n","       CYLC2    -0.016414       3.622489e-02\n","     TMEM90B     0.034482       3.622489e-02\n","      RESP18    -0.014081       3.622489e-02\n","        RGS4     0.022820       3.855158e-02\n","        CNTF    -0.020244       3.880001e-02\n","    ADAMTSL2     0.030360       4.155675e-02\n","        CST2     0.019834       4.218502e-02\n","       GRID1    -0.028724       4.218502e-02\n","       MATN3     0.028706       4.316484e-02\n","       GGT8P    -0.017000       4.494972e-02\n","      GPR101     0.014064       4.913174e-02\n","      KIF26B     0.031349       4.960098e-02\n"]},{"name":"stderr","output_type":"stream","text":["/home/techt/AI4Science_gold_standard/stage1/Haoyang/utils.py:492: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"]}],"source":["feature_cols = merged_data.columns.tolist()\n","feature_cols.remove(TRAIT)\n","\n","if has_batch_effect:\n","    report_result_from_lmm(model, feature_cols, TRAIT, condition, threshold=0.05, save_output=True,\n","                           output_dir=OUTPUT_DIR)\n","else:\n","    report_result_from_lasso(model, feature_cols, TRAIT, condition, save_output=True, output_dir=OUTPUT_DIR)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:25.000830410Z","start_time":"2023-12-31T21:20:24.982834541Z"},"id":"0VidWNDkiaED","outputId":"ae1dfede-e05a-40bf-80ca-203e5fffcef5"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":[],"metadata":{"id":"bQbxWUPBiaED"}}]}