{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"machine_shape":"hm"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Gold standard curation: Preprocessing and single-step regression"],"metadata":{"collapsed":false,"id":"NB8dJaS7ZWQI"}},{"cell_type":"markdown","source":["In this stage of gold standard curation, we will do the data preprocessing, selection, and single-step regression for the 153 traits in our question set. This file shows the reference steps using the trait \"Breast Cancer\" as an example. The workflow consists of the following steps:\n","\n","1. Preprocess all the cohorts related to this trait. Each cohort should be converted to a tabular form and saved to a csv file, with columns being genetic factors, the trait, and age, gender if available;\n","2. If there exists at least one cohort with age or gender information, conduct regression analysis with genetic features together with age or gender as the regressors.\n"],"metadata":{"collapsed":false,"id":"T8BiY3jEZWQL"}},{"cell_type":"markdown","source":["# 1. Basic setup"],"metadata":{"collapsed":false,"id":"6rY1UNA-ZWQM"}},{"cell_type":"code","execution_count":2,"outputs":[],"source":["import os\n","\n","# Set your preferred name\n","USER = \"Shuyi\"\n","# Set the data and output directories\n","DATA_ROOT = '/content/drive/MyDrive/AI4Science_Public/data'\n","OUTPUT_ROOT = '/content/drive/MyDrive/AI4Science_Public/output1'\n","TRAIT = \"Lower Grade Glioma\"\n","\n","OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n","JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n","if not os.path.exists(OUTPUT_DIR):\n","    os.makedirs(OUTPUT_DIR, exist_ok=True)\n","\n","# Gene symbol normalization may take 1-2 minutes. You may set it to False for debugging.\n","NORMALIZE_GENE = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:07.313098989Z","start_time":"2023-12-31T21:19:07.312718292Z"},"id":"CPen6psLZWQM","executionInfo":{"status":"ok","timestamp":1704460088946,"user_tz":-480,"elapsed":2369,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jMTdsU43vVg3","outputId":"3baa5819-ff36-47ed-9843-08ecc21efe83","ExecuteTime":{"end_time":"2023-12-31T21:19:07.680659936Z","start_time":"2023-12-31T21:19:07.666959145Z"},"executionInfo":{"status":"ok","timestamp":1704460062669,"user_tz":-480,"elapsed":213912,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}],"source":["# This cell is only for use on Google Colab. Skip it if you run your code in other environments\n","\n","import os\n","from google.colab import drive\n","\n","drive.mount('/content/drive', force_remount=True)\n","proj_dir = '/content/drive/MyDrive/AI4Science_Public'\n","os.chdir(proj_dir)"]},{"cell_type":"code","source":["import sys\n","sys.path.append('/content/drive/MyDrive/Colab Notebooks')\n","!pip install mygene\n","!pip install sparse_lmm"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ucJm9PCJaIf8","executionInfo":{"status":"ok","timestamp":1704460104122,"user_tz":-480,"elapsed":12952,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"b2ee29ea-273c-47f9-d481-821eb6876129"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting mygene\n","  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n","Collecting biothings-client>=0.2.6 (from mygene)\n","  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n","Requirement already satisfied: requests>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from biothings-client>=0.2.6->mygene) (2.31.0)\n","Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.3.2)\n","Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (3.6)\n","Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2.0.7)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.3.0->biothings-client>=0.2.6->mygene) (2023.11.17)\n","Installing collected packages: biothings-client, mygene\n","Successfully installed biothings-client-0.3.1 mygene-3.2.2\n","Collecting sparse_lmm\n","  Downloading sparse_lmm-0.4-py3-none-any.whl (7.1 kB)\n","Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from sparse_lmm) (1.11.4)\n","Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from sparse_lmm) (1.23.5)\n","Installing collected packages: sparse_lmm\n","Successfully installed sparse_lmm-0.4\n"]}]},{"cell_type":"markdown","source":["# 2. Data preprocessing and selection"],"metadata":{"collapsed":false,"id":"uuyW6piSZWQP"}},{"cell_type":"markdown","source":["## 2.1. The TCGA Xena dataset"],"metadata":{"collapsed":false,"id":"gpsSPrUeZWQP"}},{"cell_type":"markdown","source":["In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."],"metadata":{"collapsed":false,"id":"xRKc-zchZWQQ"}},{"cell_type":"code","execution_count":4,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['.DS_Store',\n"," 'TCGA_Esophageal_Cancer_(ESCA)',\n"," 'TCGA_Pheochromocytoma_Paraganglioma_(PCPG)',\n"," 'TCGA_Bile_Duct_Cancer_(CHOL)',\n"," 'TCGA_Kidney_Chromophobe_(KICH)',\n"," 'TCGA_Thyroid_Cancer_(THCA)',\n"," 'TCGA_Sarcoma_(SARC)',\n"," 'TCGA_Adrenocortical_Cancer_(ACC)',\n"," 'TCGA_Thymoma_(THYM)',\n"," 'TCGA_Glioblastoma_(GBM)']"]},"metadata":{},"execution_count":4}],"source":["import os\n","from utils import *\n","\n","dataset = 'TCGA'\n","dataset_dir = os.path.join(DATA_ROOT, dataset)\n","os.listdir(dataset_dir)[:10]"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.770095084Z","start_time":"2023-12-31T21:19:09.102451553Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"YPDQJCUYZWQQ","executionInfo":{"status":"ok","timestamp":1704460111199,"user_tz":-480,"elapsed":3590,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"47bae3f2-b89b-4d34-92c2-64d72a894e14"}},{"cell_type":"markdown","source":["If no match is found, jump directly to GEO in Part 2.2"],"metadata":{"collapsed":false,"id":"mTPvkvIcZWQR"}},{"cell_type":"code","execution_count":5,"outputs":[],"source":["trait_subdir = \"TCGA_Lower_Grade_Glioma_(LGG)\"\n","cohort = 'Xena'\n","# All the cancer traits in Xena are binary\n","trait_type = 'binary'\n","# Once a relevant cohort is found in Xena, we can generally assume the gene and clinical data are available\n","is_available = True\n","\n","clinical_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.LGG.sampleMap_LGG_clinicalMatrix')\n","genetic_data_file = os.path.join(dataset_dir, trait_subdir, 'TCGA.LGG.sampleMap_HiSeqV2_PANCAN.gz')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:09.918243934Z","start_time":"2023-12-31T21:19:09.909021135Z"},"id":"_7r3vQ4DZWQR","executionInfo":{"status":"ok","timestamp":1704460146940,"user_tz":-480,"elapsed":5,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","source":["import pandas as pd\n","\n","clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n","genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n","age_col = gender_col = None"],"metadata":{"id":"MudwB-_iz7sc","ExecuteTime":{"end_time":"2023-12-31T21:19:13.613198409Z","start_time":"2023-12-31T21:19:10.409688003Z"},"executionInfo":{"status":"ok","timestamp":1704460156476,"user_tz":-480,"elapsed":5037,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}},"execution_count":6,"outputs":[]},{"cell_type":"code","execution_count":7,"outputs":[],"source":["_, clinical_data_cols = check_rows_and_columns(clinical_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.617006167Z","start_time":"2023-12-31T21:19:13.614080653Z"},"id":"At6no_ODZWQR","executionInfo":{"status":"ok","timestamp":1704460159402,"user_tz":-480,"elapsed":489,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":8,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['_INTEGRATION',\n"," '_PATIENT',\n"," '_cohort',\n"," '_primary_disease',\n"," '_primary_site',\n"," 'additional_pharmaceutical_therapy',\n"," 'additional_radiation_therapy',\n"," 'additional_surgery_locoregional_procedure',\n"," 'additional_surgery_metastatic_procedure',\n"," 'age_at_initial_pathologic_diagnosis']"]},"metadata":{},"execution_count":8}],"source":["clinical_data_cols[:10]"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661538569Z","start_time":"2023-12-31T21:19:13.615946213Z"},"id":"IXOwP6DCZWQR","outputId":"f3c0ea1f-2b0a-4c68-faf1-c0641f40bb3e","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704460162296,"user_tz":-480,"elapsed":516,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-28T20:20:11.841042813Z","start_time":"2023-12-28T20:20:11.834458206Z"},"id":"i3SiZnBCZWQS"}},{"cell_type":"code","execution_count":9,"outputs":[{"output_type":"execute_result","data":{"text/plain":["\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_surgery_locoregional_procedure', 'additional_surgery_metastatic_procedure', 'age_at_initial_pathologic_diagnosis', 'animal_insect_allergy_history', 'animal_insect_allergy_types', 'asthma_history', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'days_to_additional_surgery_locoregional_procedure', 'days_to_additional_surgery_metastatic_procedure', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_after_initial_treatment', 'days_to_performance_status_assessment', 'eastern_cancer_oncology_group', 'eczema_history', 'family_history_of_cancer', 'family_history_of_primary_brain_tumor', 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust', 'first_diagnosis_age_of_animal_insect_allergy', 'first_diagnosis_age_of_food_allergy', 'first_presenting_symptom', 'first_presenting_symptom_longest_duration', 'followup_case_report_form_submission_reason', 'followup_treatment_success', 'food_allergy_history', 'food_allergy_types', 'form_completion_date', 'gender', 'hay_fever_history', 'headache_history', 'histological_type', 'history_ionizing_rt_to_head', 'history_of_neoadjuvant_treatment', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'inherited_genetic_syndrome_found', 'inherited_genetic_syndrome_result', 'initial_weight', 'intermediate_dimension', 'is_ffpe', 'karnofsky_performance_score', 'laterality', 'ldh1_mutation_found', 'ldh1_mutation_test_method', 'ldh1_mutation_tested', 'longest_dimension', 'lost_follow_up', 'mental_status_changes', 'mold_or_dust_allergy_history', 'motor_movement_changes', 'neoplasm_histologic_grade', 'new_tumor_event_after_initial_treatment', 'oct_embedded', 'other_dx', 'pathology_report_file_name', 'patient_id', 'performance_status_scale_timing', 'person_neoplasm_cancer_status', 'preoperative_antiseizure_meds', 'preoperative_corticosteroids', 'primary_therapy_outcome_success', 'radiation_therapy', 'sample_type', 'sample_type_id', 'seizure_history', 'sensory_changes', 'shortest_dimension', 'supratentorial_localization', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_location', 'tumor_tissue_site', 'vial_number', 'visual_changes', 'vital_status', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_LGG_mutation', '_GENOMIC_ID_TCGA_LGG_PDMRNAseq', '_GENOMIC_ID_TCGA_LGG_RPPA', '_GENOMIC_ID_TCGA_LGG_mutation_broad_gene', '_GENOMIC_ID_TCGA_LGG_gistic2', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_LGG_miRNA_HiSeq', '_GENOMIC_ID_TCGA_LGG_PDMarrayCNV', '_GENOMIC_ID_data/public/TCGA/LGG/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_LGG_mutation_curated_broad_gene', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2_percentile', '_GENOMIC_ID_TCGA_LGG_hMethyl450_MethylMix', '_GENOMIC_ID_TCGA_LGG_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_LGG_mutation_bcm_gene', '_GENOMIC_ID_TCGA_LGG_hMethyl450', '_GENOMIC_ID_TCGA_LGG_PDMarray', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2', '_GENOMIC_ID_TCGA_LGG_G4502A_07_3', '_GENOMIC_ID_TCGA_LGG_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_LGG_gistic2thd', '_GENOMIC_ID_TCGA_LGG_mutation_ucsc_maf_gene']\\n\""],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":9}],"source":["f'''\n","Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n","candidate_age_cols = [col_name1, col_name2, ...]\n","candidate_gender_cols = [col_name1, col_name2, ...]\n","If no columns match a criterion, please provide an empty list.\n","\n","Column names:\n","{clinical_data_cols}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661858700Z","start_time":"2023-12-31T21:19:13.656749487Z"},"id":"x8P_531oZWQS","outputId":"b99f8e7f-e60c-4240-e834-e793fc9015e7","colab":{"base_uri":"https://localhost:8080/","height":178},"executionInfo":{"status":"ok","timestamp":1704460165840,"user_tz":-480,"elapsed":482,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":12,"outputs":[],"source":["candidate_age_cols = ['age_at_initial_pathologic_diagnosis', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust', 'first_diagnosis_age_of_animal_insect_allergy', 'first_diagnosis_age_of_food_allergy', 'year_of_initial_pathologic_diagnosis']\n","\n","candidate_gender_cols = ['gender']"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.661938833Z","start_time":"2023-12-31T21:19:13.656901848Z"},"id":"2OWLgaibZWQS","executionInfo":{"status":"ok","timestamp":1704460193220,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Choose a single column from the candidate columns that record age and gender information respectively.\n","If no column meets the requirement, keep 'age_col' or 'gender_col' to None"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:57:44.207565572Z","start_time":"2023-12-31T03:57:44.202177544Z"},"id":"o29AdGTRZWQS"}},{"cell_type":"code","execution_count":13,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{'age_at_initial_pathologic_diagnosis': [31.0, 67.0, 44.0, 37.0, 50.0],\n"," 'days_to_birth': [-11509.0, -24578.0, -16297.0, -13565.0, -18494.0],\n"," 'days_to_collection': [nan, nan, nan, nan, nan],\n"," 'days_to_death': [nan, 234.0, 1335.0, 1106.0, nan],\n"," 'days_to_initial_pathologic_diagnosis': [0.0, 0.0, 0.0, 0.0, 0.0],\n"," 'days_to_last_followup': [3574.0, nan, nan, 481.0, 1828.0],\n"," 'first_diagnosis_age_asth_ecz_hay_fev_mold_dust': [nan, nan, nan, nan, nan],\n"," 'first_diagnosis_age_of_animal_insect_allergy': ['12 - 20 Years',\n","  nan,\n","  nan,\n","  nan,\n","  nan],\n"," 'first_diagnosis_age_of_food_allergy': [nan, nan, nan, nan, nan],\n"," 'year_of_initial_pathologic_diagnosis': [2005.0,\n","  2005.0,\n","  2006.0,\n","  2009.0,\n","  2010.0]}"]},"metadata":{},"execution_count":13}],"source":["preview_df(clinical_data[candidate_age_cols])"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.662078139Z","start_time":"2023-12-31T21:19:13.657006017Z"},"id":"ZES_DWC8ZWQS","outputId":"1a397575-0986-4c69-90fb-ae28125ea668","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704460206120,"user_tz":-480,"elapsed":485,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":14,"outputs":[],"source":["age_col = 'age_at_initial_pathologic_diagnosis'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:13.777159137Z","start_time":"2023-12-31T21:19:13.756210135Z"},"id":"Shy0HJpkZWQS","executionInfo":{"status":"ok","timestamp":1704460209683,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":15,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{'gender': ['FEMALE', 'MALE', 'FEMALE', 'MALE', 'MALE']}"]},"metadata":{},"execution_count":15}],"source":["preview_df(clinical_data[candidate_gender_cols])"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.313425023Z","start_time":"2023-12-31T21:19:14.305221062Z"},"id":"2dEG2RPBZWQS","outputId":"275a4463-4baa-4c8d-e081-a5e32c598f81","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704460210696,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":16,"outputs":[],"source":["gender_col = 'gender'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:14.852502962Z","start_time":"2023-12-31T21:19:14.845153121Z"},"id":"kNH4VpYKZWQT","executionInfo":{"status":"ok","timestamp":1704460212696,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":17,"outputs":[],"source":["selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col, gender_col)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.246065025Z","start_time":"2023-12-31T21:19:15.240279162Z"},"id":"AGQ0X4scZWQT","executionInfo":{"status":"ok","timestamp":1704460213907,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":18,"outputs":[{"output_type":"stream","name":"stderr","text":["INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('GTF2IP1', 2), ('RBMY1A3P', 3), ('HERC2P2', 3), ('WASH3P', 3), ('NUDT9P1', 2), ('SNORD104', 2), ('\n","WARNING:biothings.client:153 input query terms found no hit:\t['C16orf13', 'C16orf11', 'LOC100272146', 'LOC339240', 'NACAP1', 'LOC441204', 'KLRA1', 'FAM183A', 'FA\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:9 input query terms found dup hits:\t[('SUGT1P1', 2), ('SNORD127', 2), ('SNORA62', 4), ('IFITM4P', 8), ('HLA-DRB6', 2), ('FUNDC2P2', 2), \n","WARNING:biothings.client:189 input query terms found no hit:\t['NARFL', 'NFKBIL2', 'LOC150197', 'TMEM84', 'LOC162632', 'PPPDE1', 'PPPDE2', 'C1orf38', 'C1orf31', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:20 input query terms found dup hits:\t[('PIP5K1P1', 2), ('SNORA40', 2), ('SNORD116-24', 2), ('SNORD116-25', 2), ('SNORD116-23', 2), ('SNOR\n","WARNING:biothings.client:149 input query terms found no hit:\t['FAM153C', 'C9orf167', 'CLK2P', 'CCDC76', 'CCDC75', 'CCDC72', 'HIST3H2BB', 'PRAC', 'LOC285780', 'LO\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('SNORD58C', 2), ('UOX', 2), ('UBE2Q2P1', 3), ('PPP4R1L', 2), ('SNORD63', 3), ('SNORD116-22', 2), (\n","WARNING:biothings.client:157 input query terms found no hit:\t['C16orf91', 'LOC100130581', 'SFRS2', 'C16orf5', 'C16orf7', 'C16orf3', 'C21orf7', 'C21orf2', 'LOC389\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('FAM66D', 2), ('FAM66A', 2), ('THSD1P1', 2), ('MSLNL', 2), ('EEF1DP3', 2), ('PGM5P2', 2), ('UBE2MP\n","WARNING:biothings.client:168 input query terms found no hit:\t['LOC284551', 'LOC285548', 'LOC728410', 'LOC541473', 'DULLARD', 'KIAA0368', 'EFTUD1', 'TWISTNB', 'SF\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('S100A7L2', 2), ('POM121L8P', 2), ('SNORD16', 2), ('MEG8', 2), ('RPL23P8', 2), ('KIR3DX1', 5), ('R\n","WARNING:biothings.client:165 input query terms found no hit:\t['TMEM188', 'PDZD3', 'FAM102B', 'FAM102A', 'SMCR7L', 'G6PC', 'OSTCL', 'LOC653544', 'LOC653545', 'USP\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:20 input query terms found dup hits:\t[('SNORD50B', 2), ('PCNAP1', 2), ('SNORA63', 7), ('C5orf60', 2), ('CEACAM22P', 2), ('SNORA16A', 2), \n","WARNING:biothings.client:147 input query terms found no hit:\t['LRRC37A4', 'LOC100131726', 'CPSF3L', 'COL4A3BP', 'PAR1', 'LOC92973', 'MICALCL', 'SMCR7', 'HIST4H4'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('SDHAP3', 2), ('PSORS1C3', 8), ('POM121L10P', 2), ('HLA-J', 9), ('HLA-H', 9), ('HLA-L', 7), ('NME2\n","WARNING:biothings.client:152 input query terms found no hit:\t['LOC100131193', 'KIAA1109', 'DUS2L', 'MYST1', 'IGLL3', 'C12orf69', 'NEURL', 'FAM188B', 'FLJ32063', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('PLGLA', 2), ('PA2G4P4', 2), ('RPL23AP7', 2), ('ST20', 2), ('RPL23AP64', 3), ('DPY19L2P2', 3), ('F\n","WARNING:biothings.client:177 input query terms found no hit:\t['LOC283314', 'RBM9', 'C20orf118', 'LOC150786', 'C20orf112', 'C20orf111', 'C20orf117', 'C20orf114', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:19 input query terms found dup hits:\t[('TPTE2P1', 2), ('FOLH1B', 2), ('ID2B', 2), ('MSX2P1', 2), ('RAET1K', 2), ('BMS1P4', 2), ('RPL23AP8\n","WARNING:biothings.client:174 input query terms found no hit:\t['MYEOV2', 'FAM35B2', 'C20orf85', 'FAM175A', 'C17orf73', 'C17orf72', 'C17orf71', 'C17orf70', 'C17orf\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:26 input query terms found dup hits:\t[('CATSPER2P1', 2), ('AQP7P3', 2), ('AQP7P1', 2), ('ATP8B5P', 2), ('OR7E91P', 2), ('ZNF204P', 2), ('\n","WARNING:biothings.client:144 input query terms found no hit:\t['BASE', 'C14orf21', 'C14orf23', 'FAM5B', 'TROVE2', 'C18orf62', 'LOC401052', 'GNB2L1', 'FAM19A4', 'F\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('P2RX6P', 2), ('RP9P', 2), ('SNORD1A', 2), ('SNORD10', 2), ('SNORD19', 2), ('NRADDP', 2), ('AOX2P'\n","WARNING:biothings.client:168 input query terms found no hit:\t['TTC15', 'C4orf29', 'FAM48A', 'C4orf21', 'C3orf65', 'C4orf23', 'C3orf67', 'LOC144742', 'C3orf63', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:17 input query terms found dup hits:\t[('SNORA58', 2), ('PMCHL1', 2), ('FAM95B1', 2), ('TMEM191A', 2), ('RPL13P5', 2), ('SBDSP1', 2), ('CY\n","WARNING:biothings.client:143 input query terms found no hit:\t['LOC254312', 'ZNF542', 'CSDAP1', 'FAM73A', 'FAM73B', 'TCTE3', 'ZNF37B', 'LOC728875', 'KIAA1826', 'P\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('ZFP91-CNTF', 2), ('SNORD116-27', 2), ('GSTM2P1', 2), ('C3P1', 2), ('NCF1B', 2), ('SNORD114-4', 2)\n","WARNING:biothings.client:137 input query terms found no hit:\t['LOC100129716', 'C14orf184', 'C14orf181', 'C14orf183', 'C14orf182', 'C7orf26', 'C7orf27', 'C7orf23'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('ANXA2P1', 2), ('ANXA2P3', 2), ('ANXA2P2', 2), ('SNORD116-12', 2), ('ZNF286B', 2), ('SAA3P', 2), (\n","WARNING:biothings.client:132 input query terms found no hit:\t['AKR1CL1', 'CCDC21', 'CCDC23', 'TMEM111', 'APOB48R', 'LOC595101', 'KIAA1430', 'KIAA1432', 'ORC4L', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('TCP10L2', 2), ('CXADRP3', 2), ('SNORD116-13', 2), ('TPTE2P3', 2), ('SNORD116-15', 2), ('SNORD116-\n","WARNING:biothings.client:175 input query terms found no hit:\t['DFNB59', 'LOC100170939', 'FAM36A', 'C2orf70', 'C2orf71', 'C2orf77', 'C2orf79', 'LOC100128191', 'HR\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:16 input query terms found dup hits:\t[('PMS2CL', 2), ('BRD7P3', 2), ('SCARNA17', 2), ('GABARAPL3', 2), ('MT1IP', 2), ('ADAM6', 3), ('SNOR\n","WARNING:biothings.client:142 input query terms found no hit:\t['C9orf30', 'PEG3AS', 'C9orf37', 'TXNDC3', 'LOC442308', 'FAM46D', 'FAM46A', 'FAM46B', 'FAM46C', 'WDR\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:13 input query terms found dup hits:\t[('FAM197Y2', 3), ('SNORD1C', 2), ('SNORA71B', 2), ('NAT8B', 3), ('SNORA21', 2), ('VENTXP7', 2), ('G\n","WARNING:biothings.client:148 input query terms found no hit:\t['FCGR1B', 'RG9MTD2', 'RARS', 'LOC727677', 'CCRN4L', 'SEPT14', 'HKR1', 'C1orf112', 'LOC100270710', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:13 input query terms found dup hits:\t[('OR6W1P', 3), ('LYPLA2P1', 6), ('RPL23AP53', 2), ('TRPC2', 2), ('RBMY2FP', 2), ('BAGE2', 2), ('POM\n","WARNING:biothings.client:150 input query terms found no hit:\t['SIP1', 'FAM129C', 'GIF', 'TMCO7', 'CXorf61', 'CXorf64', 'ZNF192', 'SEPP1', 'KIRREL', 'MGC16121', '\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:18 input query terms found dup hits:\t[('ISCA1P1', 2), ('SNORD116-5', 2), ('PCDHB19P', 3), ('NAPSB', 2), ('WASH7P', 2), ('TTTY11', 2), ('H\n","WARNING:biothings.client:135 input query terms found no hit:\t['PPYR1', 'FAM82A2', 'FLJ37453', 'FAM82A1', 'KIAA0947', 'MOBKL2B', 'MOBKL2C', 'MOBKL2A', 'CMAH', 'GR\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-530...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('SNORD12B', 2), ('SNORD116-8', 2), ('SNORD116-4', 2), ('SNORD116-1', 2), ('SNORD116-3', 2), ('SNOR\n","WARNING:biothings.client:78 input query terms found no hit:\t['C14orf174', 'C14orf176', 'C14orf177', 'C14orf179', 'PTPLB', 'SGEF', 'LOC400752', 'GRLF1', 'SPANXE'\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"]}],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:15.593478652Z","start_time":"2023-12-31T21:19:15.583841656Z"},"id":"sxc-Bd_XZWQT","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704460269846,"user_tz":-480,"elapsed":53792,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"35988f5b-fbbd-43d1-93fc-b477838e183a"}},{"cell_type":"code","execution_count":19,"outputs":[{"output_type":"execute_result","data":{"text/plain":["                 Lower Grade Glioma   Age  Gender  ARHGEF10L     HIF3A  \\\n","sampleID                                                                 \n","TCGA-CS-4938-01                   1  31.0     0.0   1.675408  2.256574   \n","TCGA-CS-4941-01                   1  67.0     1.0   1.677108  1.531874   \n","TCGA-CS-4942-01                   1  44.0     0.0   1.451408  0.339574   \n","TCGA-CS-4943-01                   1  37.0     1.0   1.159508  1.747574   \n","TCGA-CS-4944-01                   1  50.0     1.0   0.847608  3.430774   \n","\n","                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n","sampleID                                                          ...   \n","TCGA-CS-4938-01 -0.025535 -0.129672  0.512522  1.00569  0.980606  ...   \n","TCGA-CS-4941-01 -0.103335 -0.044972  1.044022  0.68629  0.715906  ...   \n","TCGA-CS-4942-01  0.227665 -0.359972  0.479422  0.83969  0.928806  ...   \n","TCGA-CS-4943-01  0.193265  0.049128  0.345022  0.20049  0.736706  ...   \n","TCGA-CS-4944-01 -0.531035 -0.503272  0.820222  1.21089  0.934706  ...   \n","\n","                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n","sampleID                                                                     \n","TCGA-CS-4938-01  2.145814 -0.086682 -0.748878 -0.411617  0.508467 -1.28139   \n","TCGA-CS-4941-01  3.563814 -0.086682  2.294922  2.043683  0.703167 -1.28139   \n","TCGA-CS-4942-01  2.061014 -0.086682 -0.748878  2.006583  1.287867 -1.28139   \n","TCGA-CS-4943-01 -0.864186 -0.086682 -0.748878 -0.862817  0.231767 -1.28139   \n","TCGA-CS-4944-01  3.457214 -0.086682 -0.159678  1.123883  0.491667 -1.28139   \n","\n","                    TULP3     BCL6B     GSTK1      SELP  \n","sampleID                                                 \n","TCGA-CS-4938-01  0.587923 -1.931327 -1.200995 -5.525533  \n","TCGA-CS-4941-01 -0.211777 -0.691827 -0.067395 -4.274433  \n","TCGA-CS-4942-01 -0.521377 -1.996927 -1.020895 -5.096833  \n","TCGA-CS-4943-01  0.760023 -1.260727 -2.100995 -4.541333  \n","TCGA-CS-4944-01 -0.713077 -1.252627 -1.109495 -3.516133  \n","\n","[5 rows x 17350 columns]"],"text/html":["\n","  <div id=\"df-508b9dd3-d2c2-441d-833e-6546520a6ca3\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Lower Grade Glioma</th>\n","      <th>Age</th>\n","      <th>Gender</th>\n","      <th>ARHGEF10L</th>\n","      <th>HIF3A</th>\n","      <th>RNF17</th>\n","      <th>RNF10</th>\n","      <th>RNF11</th>\n","      <th>RNF13</th>\n","      <th>GTF2IP1</th>\n","      <th>...</th>\n","      <th>SLC7A10</th>\n","      <th>PLA2G2C</th>\n","      <th>TULP2</th>\n","      <th>NPY5R</th>\n","      <th>GNGT2</th>\n","      <th>GNGT1</th>\n","      <th>TULP3</th>\n","      <th>BCL6B</th>\n","      <th>GSTK1</th>\n","      <th>SELP</th>\n","    </tr>\n","    <tr>\n","      <th>sampleID</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>TCGA-CS-4938-01</th>\n","      <td>1</td>\n","      <td>31.0</td>\n","      <td>0.0</td>\n","      <td>1.675408</td>\n","      <td>2.256574</td>\n","      <td>-0.025535</td>\n","      <td>-0.129672</td>\n","      <td>0.512522</td>\n","      <td>1.00569</td>\n","      <td>0.980606</td>\n","      <td>...</td>\n","      <td>2.145814</td>\n","      <td>-0.086682</td>\n","      <td>-0.748878</td>\n","      <td>-0.411617</td>\n","      <td>0.508467</td>\n","      <td>-1.28139</td>\n","      <td>0.587923</td>\n","      <td>-1.931327</td>\n","      <td>-1.200995</td>\n","      <td>-5.525533</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4941-01</th>\n","      <td>1</td>\n","      <td>67.0</td>\n","      <td>1.0</td>\n","      <td>1.677108</td>\n","      <td>1.531874</td>\n","      <td>-0.103335</td>\n","      <td>-0.044972</td>\n","      <td>1.044022</td>\n","      <td>0.68629</td>\n","      <td>0.715906</td>\n","      <td>...</td>\n","      <td>3.563814</td>\n","      <td>-0.086682</td>\n","      <td>2.294922</td>\n","      <td>2.043683</td>\n","      <td>0.703167</td>\n","      <td>-1.28139</td>\n","      <td>-0.211777</td>\n","      <td>-0.691827</td>\n","      <td>-0.067395</td>\n","      <td>-4.274433</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4942-01</th>\n","      <td>1</td>\n","      <td>44.0</td>\n","      <td>0.0</td>\n","      <td>1.451408</td>\n","      <td>0.339574</td>\n","      <td>0.227665</td>\n","      <td>-0.359972</td>\n","      <td>0.479422</td>\n","      <td>0.83969</td>\n","      <td>0.928806</td>\n","      <td>...</td>\n","      <td>2.061014</td>\n","      <td>-0.086682</td>\n","      <td>-0.748878</td>\n","      <td>2.006583</td>\n","      <td>1.287867</td>\n","      <td>-1.28139</td>\n","      <td>-0.521377</td>\n","      <td>-1.996927</td>\n","      <td>-1.020895</td>\n","      <td>-5.096833</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4943-01</th>\n","      <td>1</td>\n","      <td>37.0</td>\n","      <td>1.0</td>\n","      <td>1.159508</td>\n","      <td>1.747574</td>\n","      <td>0.193265</td>\n","      <td>0.049128</td>\n","      <td>0.345022</td>\n","      <td>0.20049</td>\n","      <td>0.736706</td>\n","      <td>...</td>\n","      <td>-0.864186</td>\n","      <td>-0.086682</td>\n","      <td>-0.748878</td>\n","      <td>-0.862817</td>\n","      <td>0.231767</td>\n","      <td>-1.28139</td>\n","      <td>0.760023</td>\n","      <td>-1.260727</td>\n","      <td>-2.100995</td>\n","      <td>-4.541333</td>\n","    </tr>\n","    <tr>\n","      <th>TCGA-CS-4944-01</th>\n","      <td>1</td>\n","      <td>50.0</td>\n","      <td>1.0</td>\n","      <td>0.847608</td>\n","      <td>3.430774</td>\n","      <td>-0.531035</td>\n","      <td>-0.503272</td>\n","      <td>0.820222</td>\n","      <td>1.21089</td>\n","      <td>0.934706</td>\n","      <td>...</td>\n","      <td>3.457214</td>\n","      <td>-0.086682</td>\n","      <td>-0.159678</td>\n","      <td>1.123883</td>\n","      <td>0.491667</td>\n","      <td>-1.28139</td>\n","      <td>-0.713077</td>\n","      <td>-1.252627</td>\n","      <td>-1.109495</td>\n","      <td>-3.516133</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 17350 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-508b9dd3-d2c2-441d-833e-6546520a6ca3')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-508b9dd3-d2c2-441d-833e-6546520a6ca3 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-508b9dd3-d2c2-441d-833e-6546520a6ca3');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-7157ac91-ac61-4dfc-bd0f-ae2b78f7cec9\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-7157ac91-ac61-4dfc-bd0f-ae2b78f7cec9')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-7157ac91-ac61-4dfc-bd0f-ae2b78f7cec9 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":19}],"source":["merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n","merged_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.074598194Z","start_time":"2023-12-31T21:19:15.973600157Z"},"id":"52Ks-OVpZWQT","outputId":"be6f55a7-72fc-4a6f-aa62-328358327ced","colab":{"base_uri":"https://localhost:8080/","height":495},"executionInfo":{"status":"ok","timestamp":1704460275758,"user_tz":-480,"elapsed":539,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":["The merged dataset contains 529 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:16.341121723Z","start_time":"2023-12-31T21:19:16.330544222Z"},"id":"LQ-LqBo2ZWQT","outputId":"f8fdabf0-244f-452d-ebcc-9a3e67cf2f5a","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704460283099,"user_tz":-480,"elapsed":472,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":21,"outputs":[{"output_type":"stream","name":"stdout","text":["For the feature 'Lower Grade Glioma', the least common label is '1' with 529 occurrences. This represents 100.00% of the dataset.\n","The distribution of the feature 'Lower Grade Glioma' in this dataset is severely biased.\n","\n","Quartiles for 'Age':\n","  25%: 32.0\n","  50% (Median): 41.0\n","  75%: 53.0\n","Min: 14.0\n","Max: 87.0\n","The distribution of the feature 'Age' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '0.0' with 238 occurrences. This represents 44.99% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"output_type":"execute_result","data":{"text/plain":["True"]},"metadata":{},"execution_count":21}],"source":["is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:17.222505900Z","start_time":"2023-12-31T21:19:17.218781412Z"},"id":"aSrJydC2ZWQT","outputId":"8f25b9ee-49a0-48c3-d64a-7e1fea16469a","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704460285108,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":22,"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.401536463Z","start_time":"2023-12-31T21:19:17.980683628Z"},"id":"ldageITxZWQT","executionInfo":{"status":"ok","timestamp":1704460288577,"user_tz":-480,"elapsed":533,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":23,"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.405288661Z","start_time":"2023-12-31T21:19:35.402799553Z"},"id":"Kq1zPqtfZWQT","executionInfo":{"status":"ok","timestamp":1704460291034,"user_tz":-480,"elapsed":535,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["## 2.2. The GEO dataset"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-28T21:28:38.028916303Z","start_time":"2023-12-28T21:28:38.016245426Z"},"id":"zNMld4NxZWQT"}},{"cell_type":"markdown","source":["In GEO, there may be one or multiple cohorts for a trait. Each cohort is identified by an accession number. We iterate over all accession numbers in the corresponding subdirectory, preprocess the cohort data, and save them to csv files."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:25:23.253882615Z","start_time":"2023-12-31T03:25:23.244062710Z"},"id":"uGxJWgJIZWQT"}},{"cell_type":"code","execution_count":24,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['GSE74567',\n"," 'GSE4058',\n"," 'GSE28271',\n"," 'GSE143843',\n"," 'GSE35158',\n"," 'GSE111627',\n"," 'GSE19578',\n"," 'GSE107850',\n"," 'GSE24072',\n"," 'GSE14',\n"," 'GSE145510',\n"," 'GSE15309']"]},"metadata":{},"execution_count":24}],"source":["dataset = 'GEO'\n","trait_subdir = \"Lower-Grade-Glioma\"\n","\n","trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n","os.listdir(trait_path)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.411177575Z","start_time":"2023-12-31T21:19:35.405710636Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"8-VIPj3fZWQU","executionInfo":{"status":"ok","timestamp":1704460354339,"user_tz":-480,"elapsed":3835,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"a0467679-52c3-452a-c812-49730ffe0823"}},{"cell_type":"markdown","source":["Repeat the below steps for all the accession numbers"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:39:42.634870142Z","start_time":"2023-12-31T03:39:42.534093295Z"},"id":"KJEJBf79ZWQU"}},{"cell_type":"code","execution_count":145,"outputs":[{"output_type":"execute_result","data":{"text/plain":["('/content/drive/MyDrive/AI4Science_Public/data/GEO/Lower-Grade-Glioma/GSE15309/GSE15309_family.soft.gz',\n"," '/content/drive/MyDrive/AI4Science_Public/data/GEO/Lower-Grade-Glioma/GSE15309/GSE15309_series_matrix.txt.gz')"]},"metadata":{},"execution_count":145}],"source":["cohort = accession_num = \"GSE15309\"\n","cohort_dir = os.path.join(trait_path, accession_num)\n","soft_file, matrix_file = get_relevant_filepaths(cohort_dir)\n","soft_file, matrix_file"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.413836016Z","start_time":"2023-12-31T21:19:35.410443976Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"cCUVGG5KZWQU","executionInfo":{"status":"ok","timestamp":1704462306519,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"364eec15-896c-4639-a19f-a6ffc665cbe8"}},{"cell_type":"markdown","source":["### Inital filtering and clinical data preprocessing"],"metadata":{"collapsed":false,"id":"Rqk_JLieZWQU"}},{"cell_type":"code","execution_count":146,"outputs":[{"output_type":"stream","name":"stdout","text":["!Series_title\t\"Cancer Genome Anatomy Project (CGAP) Human SAGE\"\n","!Series_summary\t\"This series represents the Cancer Genome Anatomy Project SAGE library collection.  Libraries contained herein were either produced through CGAP funding, or donated to CGAP.\"\n","!Series_summary\t\"The Cancer Genome Anatomy Project (CGAP: http://cgap.nci.nih.gov) is an interdisciplinary program established and administered by the National Cancer Institute (NCI: http://www.nci.nih.gov) to generate the information and technological tools needed to decipher the molecular anatomy of the cancer cell.\"\n","!Series_summary\t\"\"\n","!Series_summary\t\"SAGE libraries are named according to the following convention:\"\n","!Series_summary\t\"* SAGE_Organ_histology_code_unique identifier, e.g., SAGE_Colon_adenocarcinoma_CL_Caco2\"\n","!Series_summary\t\"* Codes: B = bulk; CL = cell line; CS = short-term cell culture; MD = micro-dissected; AP = antibody purified.\"\n","!Series_overall_design\t\"324 libraries\"\n"]}],"source":["background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n","clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']\n","\n","background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n","print(background_info)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.492119842Z","start_time":"2023-12-31T21:19:35.413383102Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"B0jX7hOMZWQU","executionInfo":{"status":"ok","timestamp":1704462312612,"user_tz":-480,"elapsed":3801,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"43d1e102-9a66-4d68-cacb-2cc392f6f12e"}},{"cell_type":"code","execution_count":147,"outputs":[{"output_type":"execute_result","data":{"text/plain":["         !Sample_geo_accession  \\\n","0  !Sample_characteristics_ch1   \n","1  !Sample_characteristics_ch1   \n","2  !Sample_characteristics_ch1   \n","3  !Sample_characteristics_ch1   \n","4  !Sample_characteristics_ch1   \n","\n","                                        GSM383679  \\\n","0                       organ/tissue: bone marrow   \n","1                 tissue preparation: flow-sorted   \n","2                                    sex: unknown   \n","3  cell type: Stem Proginitor Cells, CD34+, CD38+   \n","4                                             NaN   \n","\n","                               GSM383680  \\\n","0              organ/tissue: bone marrow   \n","1        tissue preparation: flow-sorted   \n","2                           sex: unknown   \n","3  cell type: Acute myelogenous leukemia   \n","4                                    NaN   \n","\n","                               GSM383682  \\\n","0              organ/tissue: bone marrow   \n","1  tissue preparation: antibody purified   \n","2                           sex: unknown   \n","3          cell type: normal bone marrow   \n","4                                    NaN   \n","\n","                               GSM383683                  GSM383684  \\\n","0              organ/tissue: bone marrow  organ/tissue: bone marrow   \n","1  tissue preparation: antibody purified   tissue preparation: bulk   \n","2                           sex: unknown               sex: unknown   \n","3          cell type: normal bone marrow                        NaN   \n","4                                    NaN                        NaN   \n","\n","                                           GSM383686  \\\n","0                                organ/tissue: brain   \n","1                      tissue preparation: cell line   \n","2                                         age: fetal   \n","3  cell type: normal fetal astrocyte, GFAP positi...   \n","4  other information: Normal human astrocyte cell...   \n","\n","                                     GSM383687  \\\n","0                          organ/tissue: brain   \n","1                     tissue preparation: bulk   \n","2                                      age: 34   \n","3                                    sex: male   \n","4  cell type: pilocytic astrocyoma, cerebellum   \n","\n","                         GSM383688                        GSM383689  ...  \\\n","0              organ/tissue: brain              organ/tissue: brain  ...   \n","1         tissue preparation: bulk         tissue preparation: bulk  ...   \n","2                      age: 35 yrs                          age: 23  ...   \n","3                        sex: male                        sex: male  ...   \n","4  cell type: astrocytoma grade II  cell type: astrocytoma grade II  ...   \n","\n","                                        GSM384027  \\\n","0                 organ/tissue: white blood cells   \n","1           tissue preparation: antibody purified   \n","2                                         age: 45   \n","3                                       sex: male   \n","4  cell type: monocyte-depleted mononuclear cells   \n","\n","                                        GSM384028  \\\n","0                 organ/tissue: white blood cells   \n","1           tissue preparation: antibody purified   \n","2                                         age: 71   \n","3                                       sex: male   \n","4  cell type: monocyte-depleted mononuclear cells   \n","\n","                               GSM384029  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 45   \n","3                              sex: male   \n","4                    cell type: monocyte   \n","\n","                               GSM384030  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 71   \n","3                              sex: male   \n","4                    cell type: monocyte   \n","\n","                               GSM384031  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 39   \n","3                              sex: male   \n","4                    cell type: monocyte   \n","\n","                               GSM384032  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 68   \n","3                            sex: female   \n","4                    cell type: monocyte   \n","\n","                               GSM384033  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 72   \n","3                            sex: female   \n","4                    cell type: monocyte   \n","\n","                               GSM384034  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 31   \n","3                            sex: female   \n","4           cell type: normal leukocytes   \n","\n","                               GSM384035  \\\n","0        organ/tissue: white blood cells   \n","1  tissue preparation: antibody purified   \n","2                                age: 71   \n","3                              sex: male   \n","4           cell type: plaque macrophage   \n","\n","                               GSM384036  \n","0        organ/tissue: white blood cells  \n","1  tissue preparation: antibody purified  \n","2                                age: 72  \n","3                              sex: male  \n","4           cell type: plaque macrophage  \n","\n","[5 rows x 328 columns]"],"text/html":["\n","  <div id=\"df-3633eb79-8fb0-457e-8dec-ac6e4b409fae\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>!Sample_geo_accession</th>\n","      <th>GSM383679</th>\n","      <th>GSM383680</th>\n","      <th>GSM383682</th>\n","      <th>GSM383683</th>\n","      <th>GSM383684</th>\n","      <th>GSM383686</th>\n","      <th>GSM383687</th>\n","      <th>GSM383688</th>\n","      <th>GSM383689</th>\n","      <th>...</th>\n","      <th>GSM384027</th>\n","      <th>GSM384028</th>\n","      <th>GSM384029</th>\n","      <th>GSM384030</th>\n","      <th>GSM384031</th>\n","      <th>GSM384032</th>\n","      <th>GSM384033</th>\n","      <th>GSM384034</th>\n","      <th>GSM384035</th>\n","      <th>GSM384036</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>organ/tissue: bone marrow</td>\n","      <td>organ/tissue: bone marrow</td>\n","      <td>organ/tissue: bone marrow</td>\n","      <td>organ/tissue: bone marrow</td>\n","      <td>organ/tissue: bone marrow</td>\n","      <td>organ/tissue: brain</td>\n","      <td>organ/tissue: brain</td>\n","      <td>organ/tissue: brain</td>\n","      <td>organ/tissue: brain</td>\n","      <td>...</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","      <td>organ/tissue: white blood cells</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>tissue preparation: flow-sorted</td>\n","      <td>tissue preparation: flow-sorted</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: bulk</td>\n","      <td>tissue preparation: cell line</td>\n","      <td>tissue preparation: bulk</td>\n","      <td>tissue preparation: bulk</td>\n","      <td>tissue preparation: bulk</td>\n","      <td>...</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","      <td>tissue preparation: antibody purified</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>sex: unknown</td>\n","      <td>sex: unknown</td>\n","      <td>sex: unknown</td>\n","      <td>sex: unknown</td>\n","      <td>sex: unknown</td>\n","      <td>age: fetal</td>\n","      <td>age: 34</td>\n","      <td>age: 35 yrs</td>\n","      <td>age: 23</td>\n","      <td>...</td>\n","      <td>age: 45</td>\n","      <td>age: 71</td>\n","      <td>age: 45</td>\n","      <td>age: 71</td>\n","      <td>age: 39</td>\n","      <td>age: 68</td>\n","      <td>age: 72</td>\n","      <td>age: 31</td>\n","      <td>age: 71</td>\n","      <td>age: 72</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>cell type: Stem Proginitor Cells, CD34+, CD38+</td>\n","      <td>cell type: Acute myelogenous leukemia</td>\n","      <td>cell type: normal bone marrow</td>\n","      <td>cell type: normal bone marrow</td>\n","      <td>NaN</td>\n","      <td>cell type: normal fetal astrocyte, GFAP positi...</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","      <td>...</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","      <td>sex: female</td>\n","      <td>sex: female</td>\n","      <td>sex: female</td>\n","      <td>sex: male</td>\n","      <td>sex: male</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>!Sample_characteristics_ch1</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>other information: Normal human astrocyte cell...</td>\n","      <td>cell type: pilocytic astrocyoma, cerebellum</td>\n","      <td>cell type: astrocytoma grade II</td>\n","      <td>cell type: astrocytoma grade II</td>\n","      <td>...</td>\n","      <td>cell type: monocyte-depleted mononuclear cells</td>\n","      <td>cell type: monocyte-depleted mononuclear cells</td>\n","      <td>cell type: monocyte</td>\n","      <td>cell type: monocyte</td>\n","      <td>cell type: monocyte</td>\n","      <td>cell type: monocyte</td>\n","      <td>cell type: monocyte</td>\n","      <td>cell type: normal leukocytes</td>\n","      <td>cell type: plaque macrophage</td>\n","      <td>cell type: plaque macrophage</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 328 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3633eb79-8fb0-457e-8dec-ac6e4b409fae')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-3633eb79-8fb0-457e-8dec-ac6e4b409fae button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-3633eb79-8fb0-457e-8dec-ac6e4b409fae');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-bb7897a0-dbc7-401a-b386-86f975f0feb9\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-bb7897a0-dbc7-401a-b386-86f975f0feb9')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-bb7897a0-dbc7-401a-b386-86f975f0feb9 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":147}],"source":["clinical_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.507840422Z","start_time":"2023-12-31T21:19:35.451789107Z"},"colab":{"base_uri":"https://localhost:8080/","height":533},"id":"_oQdQ_4PZWQU","executionInfo":{"status":"ok","timestamp":1704462315257,"user_tz":-480,"elapsed":506,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"da45a4c7-f054-4bbd-9725-21e67a91e97d"}},{"cell_type":"code","execution_count":148,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{0: ['organ/tissue: bone marrow',\n","  'organ/tissue: brain',\n","  'organ/tissue: cerebellum',\n","  'organ/tissue: mammary gland',\n","  'organ/tissue: cartilage',\n","  'organ/tissue: colon',\n","  'organ/tissue: stem cell',\n","  'organ/tissue: other',\n","  'organ/tissue: eye',\n","  'organ/tissue: uncharacterized tissue',\n","  'organ/tissue: heart',\n","  'organ/tissue: kidney',\n","  'organ/tissue: white blood cells',\n","  'organ/tissue: liver',\n","  'organ/tissue: lung',\n","  'organ/tissue: lymph node',\n","  'organ/tissue: muscle',\n","  'organ/tissue: ovary',\n","  'organ/tissue: pancreas',\n","  'organ/tissue: bone',\n","  'organ/tissue: peritoneum',\n","  'organ/tissue: placenta',\n","  'organ/tissue: prostate',\n","  'organ/tissue: retina',\n","  'organ/tissue: skin',\n","  'organ/tissue: spinal cord',\n","  'organ/tissue: stomach',\n","  'organ/tissue: gastrointestinal tract',\n","  'organ/tissue: testis',\n","  'organ/tissue: thyroid',\n","  'organ/tissue: uterus',\n","  'organ/tissue: vascular'],\n"," 1: ['tissue preparation: flow-sorted',\n","  'tissue preparation: antibody purified',\n","  'tissue preparation: bulk',\n","  'tissue preparation: cell line',\n","  'tissue preparation: short term culture',\n","  'tissue preparation: microscope dissected'],\n"," 2: ['sex: unknown',\n","  'age: fetal',\n","  'age: 34',\n","  'age: 35 yrs',\n","  'age: 23',\n","  'age: 9',\n","  'age: 10 yrs',\n","  'age: 53 yrs',\n","  'age: 5 yrs',\n","  'age: 66 yrs',\n","  'age: 5',\n","  'age: 30 yrs',\n","  'age: 13 yrs',\n","  'age: 65 yrs',\n","  'age: 38 yrs',\n","  'age: 37',\n","  'age: 43 yrs old',\n","  'age: 51 yr',\n","  'age: 71 yrs',\n","  'age: 32 yrs',\n","  'age: 9 months',\n","  'age: 4 years',\n","  'age: 6.3 years',\n","  'age: 29 yrs',\n","  'age: 7',\n","  'age: 27',\n","  'age: 17',\n","  'age: 1.7 yrs',\n","  'age: 7 yrs',\n","  'age: 31',\n","  'age: 3',\n","  'age: Spontaneously aborted male/female Caucasian fetuses, ages 20-33 weeks',\n","  'age: 4 mo',\n","  'age: 51',\n","  'age: 52 yrs',\n","  'age: 80 yrs',\n","  'age: 78 yrs',\n","  'cell type: glioblastoma,  cerebral cortex',\n","  'age: 67 yrs',\n","  'age: unknown',\n","  'age: 65',\n","  'cell type: glioblatastoma infected with  mutant EGFR, cerebral cortex',\n","  'cell type: glioblatastoma infected with LacZ control, cerebral cortex',\n","  'cell type: glioblastoma, cerebral cortex',\n","  'age: 63',\n","  'cell type: medulloblastoma, cerebellum',\n","  'age: 6',\n","  'age: 4',\n","  'age: 3.9',\n","  'age: 19',\n","  'age: 9 yrs',\n","  'age: 6 yrs',\n","  'age: 2.1 yrs',\n","  'age: 3.1 yrs',\n","  'age: 22',\n","  'age: 12 yrs',\n","  'age: 28',\n","  'age: 82',\n","  'cell type: normal, cerebral cortex',\n","  'sex: male',\n","  'age: 50',\n","  'age: 42',\n","  'sex: female',\n","  'age: 29',\n","  'age: 47',\n","  'age: 56.5',\n","  'age: 52.8',\n","  'age: 44 yrs',\n","  'age: 32',\n","  'age: 41',\n","  'age: 55',\n","  'age: 67',\n","  'age: 46',\n","  'age: 11',\n","  'age: 43',\n","  'age: 35',\n","  'age: 58',\n","  'age: 44',\n","  'age: 52',\n","  'age: 53',\n","  'age: 35 years old',\n","  'age: 28 years old',\n","  'age: 68',\n","  'age: 45 years old',\n","  'cell type: adenocarcinoma',\n","  'cell type: normal epithelium',\n","  'age: Passage 20 after derivation',\n","  'age: p31',\n","  'age: p54',\n","  'age: p22',\n","  'age: passage 22',\n","  'age: passage 33',\n","  'age: passage 38',\n","  'age: passage 16',\n","  'age: p36',\n","  'age: passage 50',\n","  'age: 69',\n","  'age: N/A',\n","  'age: 60',\n","  'age: 56'],\n"," 3: ['cell type: Stem Proginitor Cells, CD34+, CD38+',\n","  'cell type: Acute myelogenous leukemia',\n","  'cell type: normal bone marrow',\n","  nan,\n","  'cell type: normal fetal astrocyte, GFAP positive, cerebral cortex',\n","  'sex: male',\n","  'sex: female',\n","  'cell type: astrocytoma, right parietal occipital lobe, cerebral cortex',\n","  'sex: unknown',\n","  'other information: 5 pooled Duke glioblastoma multiforme primary tumors.',\n","  'other information: Brain, glioblastoma multiforme cell line, H54, that has been stably transfected with epidermal growth factor receptor vIII mutant',\n","  'other information: Brain, glioblastoma multiforme cell line, H54, that has been stably transfected with the b-galactosidase gene',\n","  'other information: negative for genomic amplification of EFGR, c-myc, n-myc and gli-1',\n","  'other information: medulloblastoma grade IV',\n","  'cell type: medulloblastoma cell line',\n","  'cell type: medulloblastoma, cerebellum',\n","  'cell type: benign meningioma of the meningothelial subtype',\n","  'cell type: atypical meningioma',\n","  'cell type: Malignant Meningioma',\n","  'cell type: normal cerebellum',\n","  'other information: Brain, normal, pooled. TB-1017 derived from a 51 yo male and TB-1060 derived from a 17 yo male.',\n","  'cell type: arachnoid cells, blood vessels, macrophages',\n","  'cell type: normal pediatric cortex',\n","  'other information: Short tags are extracted from the long SAGE data',\n","  'cell type: normal thalamus',\n","  'cell type: breast  carcinoma myoepithelium',\n","  'cell type: Grade II, ER+, PR+, Her2- tumor',\n","  'cell type: Grade I, ER+, PR+, Her2- invasive ductal',\n","  'cell type: carcinoma',\n","  'cell type: breast invasive ductal carcinoma',\n","  'cell type: breast carcinoma',\n","  'cell type: ER neagtive breast cancer cell line',\n","  'cell type: ER negative breast cancer cell line',\n","  'cell type: breast carcinoma cell line',\n","  'cell type: breast  carcinoma epithelium',\n","  'cell type: Grade I, ER+, PR+, Her2- invasive ductal carcinoma',\n","  'cell type: MUC1 bead purified cells from MCF10DCIS xenografts',\n","  'cell type: ITGB6 bead purified cells from MCF10DCIS xenografts',\n","  'cell type: estrogen responsive cells purified from normal mammary epithelium',\n","  'cell type: spontenously immortalized mammary myoepithelial cell line from Li-Fraumeni patient',\n","  'other information: Colon, primary tumor',\n","  'cell type: adenocarcinoma',\n","  'mutations: Wild type p53; RER+',\n","  'mutations: RER- with mutant p53 (248arg-trp) mutant APC',\n","  'cell type: cell line derived from colorectal carcinoma',\n","  'other information: Normal colonic epithelium',\n","  'cell type: normal',\n","  'other information: Kidney, embryonic cell line 293, induced cells, expressing beta-catenin',\n","  'other information: Uninduced 293 cells',\n","  'cell type: normal bulk liver',\n","  'cell type: poorly differentiated adenocarcinoma',\n","  'cell type: poor differentiated adenocarcinoma with lymphoplasmatic infiltration',\n","  'cell type: well differentiated adenocarcinoma with lymphoplasmatic infiltration',\n","  'cell type: normal muscle biopsy of vastus lateralis',\n","  'other information: Needle muscle biopsies were obtained from the vastus lateralis muscles of 8 healthy  young men (21-24 yr) under standard conditions. All were screened fro good health and normal neuromuscular function by a panel of laboratory tests and examination by neuromuscular disease specialist. None were involved in weight lifting or extreme levels of aerobic exercise. They were asked to refrain from activities more strenuous than walking for 3 days before the biopsy. They stayed at the University of Rochester GCRC the night before the biopsy, where they received a standard meal. They received a standard breakfast about 90 min before the biopsy. The tissue was removed whitin a few minutes of administration of the local anesthetic (lidocaine). It was quickly frozen in liquid nitrogen, then stored at -70 deg C until RNA extraction.',\n","  'other information: Ovary, primary ovarian tumour, serous adenocarcinoma.',\n","  'other information: Ovary, primary tumor. Serous adenocarcinoma.',\n","  'other information: Ovary, primary ovarian tumour. Serous adenocarcinoma.',\n","  'other information: Ovary, ovarian cancer cell line A2780',\n","  'other information: Ovarian Clear cell carcinoma cell line ES-2, poorly differentiated',\n","  'cell type: cystadenoma immortalized with SV40',\n","  'other information: Ovary, surface epithelium line',\n","  'cell type: normal surface epithelium',\n","  'other information: A frozen tumor sample was obtained from a colon cancer patient. Histopathology confirmed that the tumor sample contained at least 70% of tumor cells.',\n","  'other information: ATCC Library constructed in the Kern laboratory',\n","  'cell type: normal ductal',\n","  'cell type: malignant peritoneal mesothelioma',\n","  'cell type: normal first trimester placenta',\n","  'mutations: Inactivating mutation in PTEN gene',\n","  'cell type: carcinoma, prostate glands and stroma',\n","  'other information: Prostate, cancer cell line. Induced with synthetic androgen (R1881).',\n","  'other information: LNCaP cells were maintained in RPMI 1640 (Life Technologies, Inc., Gaithersburg, MD) supplemented with 10% fetal bovine serum (FBS, Life Technologies, Inc., Gaithersburg, MD) and experiments were performed on cells between passages 20 and 30. For the studies of androgen regulation, charcoal/dextran stripped androgen-free FBS (cFBS, Gemini Bio-Products, Inc., Calabasas, CA) was used. LNCaP cells were cultured first in RPMI 1640 with 10% cFBS for 5 days and then stimulated with an non-metabolizable androgen analog R1881 (DUPONT, Boston, MA) 10-8 M for 24 hours (for generating LNCaP-T library). LNCaP cells identically treated but without R1881 treatment served as control (for generating LNCaP-C library). Cells were harvested at indicated time and polyA+ RNA was double-selected with Fast Track kit (Invitrogene) SAGE libraries were generated according to the procedure provided by Dr. Kinzler (Johns Hopkins University School of Medicine, Baltimore, MD).',\n","  'other information: LNCaP cells were maintained in RPMI 1640 supplemented with 10% fetal bovine serum  and experiments were performed on cells between passages 20 and 30. For the studies of androgen regulation, charcoal/dextran stripped androgen-free FBS  was used. LNCaP cells were cultured first in RPMI 1640 with 10% cFBS for 5 days and then stimulated with an non-metabolizable androgen analog R1881 (DUPONT, Boston, MA) 10-8 M for 24 hours (for generating LNCaP-T library). LNCaP cells identically treated but without R1881 treatment served as control (for generating LNCaP-C library). Cells were harvested at indicated time and polyA+ RNA was double-selected with Fast Track kit (Invitrogene) SAGE libraries were generated according to the procedure provided by Dr. Kinzler (Johns Hopkins University School of Medicine, Baltimore, MD).',\n","  'cell type: prostate carcinoma',\n","  'cell type: normal epithelium & stroma',\n","  'cell type: normal epithelium',\n","  'sex: male and female',\n","  'cell type: synovial fibroblasts from rheumatoid arthritis',\n","  'other information: Library made from Poly A RNA supplied by Clonetech catalog #6573-1, and LOT #0031205.',\n","  'cell type: gastric cancer',\n","  'cell type: A pool of normal gastric epithelial tissues obtained from the antrum',\n","  'cell type: normal gastric epithelium',\n","  'cell type: mixture of human cancer cell lines',\n","  'cell type: high-grade comedo DCIS endothelium',\n","  'cell type: endothelium'],\n"," 4: [nan,\n","  'other information: Normal human astrocyte cells selected from the whole brain of a 22.5 wks old donor grown in the supplied media and harvested at passage 5. Cat# CC-2565, Lot 17657.',\n","  'cell type: pilocytic astrocyoma, cerebellum',\n","  'cell type: astrocytoma grade II',\n","  'cell type: Anaplastic GradeII Astrocytoma',\n","  'cell type: anaplastic astrocytoma',\n","  'cell type: Anaplastic Astrocytoma Grade III',\n","  'cell type: anaplastic gradeIII, primary, brain',\n","  'cell type: Anaplastic GradeIII Astrocytoma',\n","  'cell type: anaplastic astrocytoma grade III,',\n","  'cell type: Anaplastic astrocytoma grade III',\n","  'cell type: ependymoma',\n","  'other information: Short tags are extracted from Long SAGE data',\n","  'cell type: pediatric glioblastoma, cerebral cortex',\n","  'cell type: glioblastoma, Lt. Temporal, cerebral cortex',\n","  'cell type: Glioblastoma',\n","  'mutations: unknown',\n","  'cell type: glioblastoma, cerebral cortex',\n","  'cell type: GBM',\n","  'cell type: medulloblastoma, cerebellum',\n","  'cell type: medulloblastoma',\n","  'cell type: medulloblastoma with neuroblastic differentiation',\n","  'cell type: desmoplastic medulloblastoma',\n","  'cell type: large cell medulloblastoma',\n","  'cell type: anaplastic medulloblastoma',\n","  'cell type: medulloblastoma,',\n","  'cell type: Medulloblastoma Cell Line',\n","  'other information: Medulloblatoma cell line treated with Otx2 siRNA.',\n","  'mutations: c-myc oncogene genomic amplification',\n","  'mutations: has pten mutation, negative on screens for 17p loss, negative for c-myc or n-myc amplification',\n","  'cell type: Medulloblastoma',\n","  'cell type: normal cerebellum',\n","  'cell type: normal whitematter, cerebral cortex',\n","  'other information: autopsy leptomeninges',\n","  'other information: Normal frontal cortex, bulk tissue, from 15 month male seizure patient (partial lobectomy).',\n","  'cell type: oligodendroglioma',\n","  'cell type: well-differentiated oligodendroglioma, frontal lobe, cortex',\n","  'other information: Extensive high-grade comedo DCIS (~8-10cm). Purified using anti-CD10 antibody and magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-MYOEP-6 myoepithelial cells\"',\n","  'cell type: breast  carcinoma myoepithelium',\n","  'cell type: ER+, PR+, HER2-, grade II',\n","  'other information: CD10+ purified myofibroblasts from invasive breast cancer. Library name in Cancer Cell, 2004, 6(1):17-32 \"I-MYOFIB-8 myofibroblasts\"',\n","  'other information: CD10+ purified myofibroblasts from invasive breast cancer  Library name in Cancer Cell, 2004, 6(1):17-32 \"I-MYOFIB-9-myofibroblasts\"',\n","  'cell type: ductal carcinoma in situ, extensive, gradeIII, Her2+',\n","  'cell type: invasive breast cancer ER+, PR+, Her2-, grade II',\n","  'cell type: carcinoma',\n","  'cell type: primary ductal in situ breast carcinoma, intermediate grade',\n","  'cell type: primary grade III. ductal invasive breast carcinoma',\n","  'other information: Ductal carcinoma in situ, high-grade, comedo.',\n","  'other information: Invasive ductal carcinoma with some in situ component, low-grade, ER-positive, erbB22-negative, lymph node-negative, p53-negative.',\n","  'other information: Breast, invasive ductal carcinoma, high-grade, ERneg, erbB2neg., lymph node neg., p53pos.',\n","  'other information: Breast, invasive ductal carcinoma, low-grade, ERpos, erbB2neg., lymph node neg., p53neg.',\n","  'other information: Estrogen receptor positive MCF7 cells were grown in conditions devoid of estrogen for 48h (i.e. IMEM media supplemented with 10% Charcoal Stripped Serum). This SAGE library represents the 0h time point (i.e. baseline, untreated cells) of an estrogen exposure time course experiment.',\n","  'other information: cells were collected 24 hours after expression of HIN-1',\n","  'other information: cells were collected 48 hours after expression of HIN-1',\n","  'other information: ZR75-1 estrogen receptor positive, estrogen dependent human breast cancer cell line, estrogen deprived for a week, cells were collected 16 hours following addition of fresh medium containing 10 nM estrogen.',\n","  'other information: ZR75-1 estrogen receptor positive, estrogen dependent human breast cancer cell line, estrogen deprived for a week, cells were collected 16 hours following addition of fresh medium containing 10 uM 4-OH-tamoxifen.',\n","  'other information: ZR75-1 human estrogen receptor positive, estrogen dependent breast cancer cell line, estrogen deprived for 1 week, cells collected 16 hours following change to fresh medium (no estrogen added).',\n","  'other information: Breast, DCIS, BerEp4 purified cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-2 epithelial cells\"',\n","  'other information: Extensive high-grade comedo DCIS (~8-10 cm). Purified using BerEP4 beads (Dynal).Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-6 epithelial cells\"',\n","  'cell type: breast  carcinoma epithelium',\n","  'other information: BerEP4 purified epithelial cells from invasive breast cancer. Library name in Cancer Cell, 2004, 6(1):17-32 \"I-EPI-8 epithelial cells\"',\n","  'other information: BerEP4 purified epithelial cells from invasive breast cancer. Library name in Cancer Cell, 2004, 6(1):17-32 \"I-EPI-9 epithelial cells\"',\n","  'cell type: invasive breast cancer, ER+, PR+, Her-, grade II',\n","  'cell type: Invasive ductal carcinoma, ER+,PR+,HER2-',\n","  'cell type: Invasive ductal carcinoma, high-grade, ER+,PR+,HER2-',\n","  'cell type: Pleural effusion, HER2-, recurrence of ER+ tumor',\n","  'cell type: Ascites, mixed lobular/ductal carcinoma, HER2-, recurrence of ER+ tumor',\n","  'cell type: invasive ductal carcinoma ER+,PR+,HER2-',\n","  'mutations: 8q gain, CDKN1A and LRP1B homozygous deletions',\n","  'cell type: ductal carcinoma in situ',\n","  'cell type: extensive LCIS',\n","  'cell type: breast carcinoma metastasis to lung',\n","  'cell type: juvenile fibroadenoma',\n","  'cell type: metastasis from breast tumor',\n","  'cell type: breast hyperplasia',\n","  'cell type: normal epithelium',\n","  'cell type: normal',\n","  'cell type: Normal breast tissue reduction',\n","  'cell type: normal breast tissue reduction',\n","  'other information: cells purified using MUC1 antibody, these cells are MUC1+ mammary epithelial cells',\n","  'cell type: normal breast tissue from a breast cancer patient (corresponding to IDC7)',\n","  'cell type: breast myoepithelium',\n","  'mutations: Li-Fraumeni patient (heterozygous TP53 mutation)',\n","  'cell type: normal breast organoid',\n","  'cell type: normal breast stroma',\n","  'cell type: recurrent phyllodes tumor, malignant, high grade, poorly differentiated',\n","  'cell type: chondrosarcoma grade 2.',\n","  'cell type: Dedifferentiated chondrosarcoma lung metastasis',\n","  'mutations: RER- (replication error negitive cell line), has normal mismatch repair',\n","  'mutations: Ras gene codon 13; wild type p53; RER+',\n","  'other information: Colon, cancer cell line.',\n","  'other information: Colon, cancer cell line. Mismatch proficient(RER-) with a mutant p53(248arg -> trp) and a mutant APC.',\n","  'cell type: adenocarcinoma',\n","  'cell type: high grade dysplasia',\n","  'cell type: low grade dysplasia',\n","  'mutations: N/A',\n","  'other information: SV-40 treated fibroblasts grown in culture after crisis in which the ALT pathway of telomerase maintenance is used',\n","  'cell type: Adenocarcinoma',\n","  'cell type: Adenocarcinoma, tubular, poorly differentiated',\n","  'cell type: Adenocarcinoma, tubular'],\n"," 5: [nan,\n","  'other information: grade one(1) pilocytic astrocytoma',\n","  'other information: Malignant embryonal tumor consistent with ependymoblastoma, in patient who subsequently had recurrent ependymal tumors',\n","  'other information: recurrent glioma from patient with previous ependymoblastoma in same fronto-parietal region',\n","  'other information: recurrent Neuroectodermal tumor w/epend diff. tumor, from 6.3 year old female with history of ependymoma',\n","  'other information: primary ependymoma in 34 yo female',\n","  'other information: primary brain tumor in right foramen of lushka',\n","  'other information: Grade II',\n","  'other information: Suboccipital Craniectomy',\n","  'other information: Lt. parietal occipital infantile CNS tumor consistent with glioblastoma multiforme',\n","  'mutations: No 17p loss or p53 mutations detected, Rasheed et al Can. Res. 54 p1324',\n","  'other information: Brain, c-myc positive medulloblastoma cell line H341',\n","  'other information: Brain, cell line mhh-1',\n","  'other information: Brain, normal, greater than 95% white matter. Post-mortem delay of 3:15. 82 year old male.',\n","  'other information: Tumor grade 2/4',\n","  'other information: Brain, well differentiated oligodendroglioma',\n","  'other information: Invasive breast tumor (~2cm), extensive high-grade comedo DCIS (~4 cm), positive lymph nodes. Purified using anti-CD10 antibody and magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-MYOEP-7 myoepithelial cells\"',\n","  'mutations: ER-, PR-',\n","  'mutations: ER+, PR+',\n","  'other information: 1. Short SAGE tags are extracted from Long SAGE data.    2. Estrogen and progesteron receptor positive, erbB2 negative, 0/25 lymph nodes (no lymph node metastasis)',\n","  'other information: Short tags are extracted from the long SAGE data',\n","  'other information: Ductal carcinoma in situ, high-grade, comedo from 32 yo woman.',\n","  'other information: Breast, ductal carcinoma in situ, high-grade, comedo, Ber-EP4 purified cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-3 epithelial cells\"',\n","  'other information: Invasive breast tumor(~2cm), extensive high-grade comedo DCIS (~4 cm), positive lymph nodes. Purified using BerEP4 beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-7 epithelial cells\"',\n","  'other information: BerEP4 purified epithelial cells',\n","  'other information: cells purified using CD24 antibody, these cells are CD24+ more differentiated breast cancer cells',\n","  'other information: cells purified using CD44 antibody, these cells are CD44+ breast \"cancer stem\" cells',\n","  'other information: cells purified using PROCR antibody, these cells are CD44+ breast \"cancer stem\" cells',\n","  'other information: Breast, macroscopically dissected (>90% pure)  high grade, comedo DCIS (ductal carcinoma in situ) of the breast obtained from a 67 year old female patient undergoing partial mastectomy. The tumor cells are estrogen and progesteron receptor positive',\n","  'mutations: Grade III, ERpos, PRlow, Her2neg',\n","  'other information: Library name in Cancer Cell, 2004, 6(1):17-32 \"FA\"',\n","  'other information: Metastatic tumor (lymph node metastasis from breast primary tumor). Age at discovery of primary tumor: 56.5 years, 10 postive lymph nodes, Primary tumor is 3 cm, histo grade = 3, nuclear grade = 3, ER-, PR-.',\n","  'other information: Metastatic tumor (lymph node metastasis from breast primary tumor). Age at discovery of primary tumor: 52.8 years, 16 postive lymph nodes.  Primary tumor is 3.2 cm, histo grade = 3, nuclear grade = 3, ER+, PR+.',\n","  'other information: normal mammary epithelium (gestational hyperplasia) from a 31 year old breast cancer patient (20 weeks pregnant).',\n","  'other information: Normal human luminal mammary epithelial cells purified with Epithelial Enrich (BER-EP4 antibody conjugated to magnetic beads) tissue obtained from a 43 year old female patient undergoing reduction mammoplasty. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-EPI-1 epithelial cells\"',\n","  'other information: Breast, normal luminal mammary epithelium, BerEp4 purified cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"n-EPI-2 epithelial cells\"',\n","  'other information: cells purified using CD24 antibody, these cells are CD24+ mammary epithelial cells',\n","  'other information: cells purified using CD24 antibody, these cells are CD24+ differentiated luminal epithelial cells',\n","  'other information: cells purified using CD44 antibody, these cells are CD44+ mammary epithelial stem cells',\n","  'other information: cells purified using MUC1 antibody, these cells are MUC1+ mammary epithelial cells',\n","  'other information: CD10 positive purified myoepithelial cells. Short tags extracted from long. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-MYO-I7 myoepithelial cells\"',\n","  'other information: normal mammary myoepithelial cells from breast reduction tissue of a 27 year old female. Antibody purified with CD10 to obtain myoepithelial cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-MYOEP-1 myoepithelial cells\"',\n","  'other information: cells purified using CD10 antibody, these cells are CD10+ mammary myoepithelial cells',\n","  'other information: Normal breast organoids from reduction mammoplasty.',\n","  'other information: Normal breast tissue from reduction mammoplasty. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-STR-1-fibroblasts\"',\n","  'other information: Library name in Cancer Cell, 2004, 6(1):17-32 \"PHY\"',\n","  'mutations: N/A',\n","  'other information: Colon, colorectal carcinoma cell line (RER-), ATCC: HTB-37.',\n","  'other information: cell line derived from colorectal carcinoma. ATCC: CCL-247. Mutation in the Ras gene, codon 13. Wild type p53. RER+',\n","  'other information: Total RNA from Ambion, Cat #7966, Lot #051P43C',\n","  'other information: Source:  Clontech Total RNA, Cat.# 64012-1, Lot# 2120258.  Right kidney; poorly differentiated with squamous differentiation (grade III-IV); extensive invasion seen in kidney, including renal tubules, lymphovascular areas and perinephric fat; no evidence of tumor in ureter, artery or vein.',\n","  'other information: Normal kidney female. Cause of death IC brain hemorrhage. Poly A RNA from Ambion Cat#7977 lot #100P94B',\n","  'other information: Pooled Femal Adult total RNA from Stratagene, cat#:735016, lot#0110204',\n","  'other information: Ambion cat #7968, lot #011P27A. Whole lung RNA from motor vehicle accident victim.',\n","  'other information: Full-term placenta.',\n","  'other information: PC3 cells, AS2 treatment, total number of 42,236 SAGE tags sequenced. 40,768 SAGE tags analyzed after clean-up (extraction of poly(A) tags, linker-generated tags, and tags originated from clones producing 4 tags or less).',\n","  'other information: Libraries were constructed using RNA isolated from pooled 4 mm diameter punches of the macula (encompassing the fovea) and matched 4 mm diameter punches of pooled mid-peripheral retina from the same five donors designated, \"4Mac\" and \"4Peri\", respectively.',\n","  'other information: The library was constructed using RNA isolated from pooled 4 mm punches of human macula and central peripheral retina from 7 morphologically normal donor eyes.',\n","  'other information: Patient had periocular malignancies that required exenteration of a normal eye as part of the clinically indicated surgical therapy. The eye was placed in saline after it was removed from the patient and peripheral retinal fragments were collected from regions outside the central arcades and frozen within 45 minutes of the surgical removal to ensure minimal degradation of mRNA molecules.',\n","  'other information: A normal peripheral retina from a 44-year-old male. Patient had periocular malignancies that required exenteration of a normal eye as part of the clinically indicated surgical therapy. The eye was placed in saline after it was removed from the patient and peripheral retinal fragments were collected from regions outside the central arcades and frozen within 45 minutes of the surgical removal to ensure minimal degradation of mRNA molecules.',\n","  'other information: RPE/choroid libraries were made from 4 mm diameter punches of macular RPE/choroid (4MacRPE) and matching 4 mm diameter punches of mid-peripheral RPE/choroid, (4PeriRPE), taken from five donor eyes.  While these libraries were made from RPE/choroid, subsequent analysis has revealed significant content of retinal transcripts in addition to RPE markers.',\n","  'other information: Gastroesophageal junction adenocarcinoma, T4N0M0, moderately differentiated, obtained by surgery.',\n","  'other information: lower esophagus close to gastroesophageal junction adenocarcinoma, T4N0M0, moderately differentiated, obtained by surgery.',\n","  'other information: Gastroesophageal junction adenocarcinoma, T4N0M0, moderately differentiated, obtained by surgery. Xenograft.',\n","  'other information: Normal breast tissue from reduction mammoplasty. Purified using anti-P1H12 antibody and magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"N-END-1-endothelial cells\"',\n","  'other information: White blood cells, isolated from a mammary gland ductal carcinoma. Purified using CD45+CD15 coupled  magnetic beads (Dynal). Extensive high-grade comedo DCIS (~8-10cm). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-LEU-6-leukocytes\"',\n","  'other information: Library name in Cancer Cell, 2004, 6(1):17-32 \"D-LEU-7-leukocytes\"',\n","  'other information: CD45+ purified tumor  infiltrating leukocytes',\n","  'mutations: none',\n","  'other information: White blood cells isolated from normal breast tissue from reduction mammoplasty. Purified using CD45+CD15 coupled magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"N-LEU-1-leukocytes\"'],\n"," 6: [nan,\n","  'other information: Invasive ductal carcinoma - primary tumor.  Age at discovery 56.5 years, 10 postive lymph nodes, 3 cm primary tumor, histo grade = 3, nuclear grade = 3, ER-, PR-.',\n","  'other information: Invasive ductal carcinoma - primary tumor.  Age at discovery 52.8 years, 16 postive lymph nodes, 3.2 cm primary tumor, histo grade = 3, nuclear grade = 3, ER+, PR+.',\n","  'other information: Lung metastasis of a breast cancer.',\n","  'other information: Tissue section of left pelvis from 35 years old male with grade 2 recurrent chondrosarcoma was used for mRNA isolation.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of nterior medial aspect of the pelvis from 28 years old male with grade 2 recurrent chondrosarcoma was divided, by gross analysis, into a cartilage-like sample and was used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in growth medium that contains heat inactivated FBS (Life Technologies Gibco) (12%), DMEM, glucose (4.5 gm/l), HEPES buffer (25 mM) solution and gentamicin sulfate (50 â▌g/ml) (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of nterior medial aspect of the pelvis from 28 years old male with grade 2 recurrent chondrosarcoma was divided, by gross analysis, into a fibrosis-like sample and was used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in growth medium that contains heat inactivated FBS (Life Technologies Gibco) (12%), DMEM, glucose (4.5 gm/l), HEPES buffer (25 mM) solution and gentamicin sulfate (50 â▌g/ml) (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of lung right lower lobe, middle posterior metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of lung right middle lobe, upper anterior   metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of lung right lower lobe, middle diaphragmatic metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of lung right upper lobe, apical lateral metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.',\n","  'other information: Tissue section of lung left lower lobe, diaphragmatic anterior metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.']}"]},"metadata":{},"execution_count":148}],"source":["clinical_data_unique = get_unique_values_by_row(clinical_data)\n","clinical_data_unique"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508015426Z","start_time":"2023-12-31T21:19:35.492746637Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"jym8RXfxZWQU","executionInfo":{"status":"ok","timestamp":1704462318549,"user_tz":-480,"elapsed":521,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"d13501dc-4967-4bdf-bb2e-c5508b2e2132"}},{"cell_type":"markdown","source":["Analyze the metadata to determine data relevance and find ways to extract the clinical data.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978204446Z","start_time":"2023-12-31T03:58:04.922270095Z"},"id":"a9Pe1BKHZWQU"}},{"cell_type":"code","execution_count":149,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'As a biomedical research team, we are selecting datasets to study the association between the human trait \\'Lower Grade Glioma\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\\n1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\\n2. For each of the traits \\'Lower Grade Glioma\\', \\'age\\', and \\'gender\\', please address these points:\\n   (1) Is there human data available for this trait?\\n   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\\n   (3) Choose an appropriate data type (either \\'continuous\\' or \\'binary\\') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\\n   Name the functions \\'convert_trait\\', \\'convert_age\\', and \\'convert_gender\\', respectively.\\n\\nBackground information about the dataset:\\n!Series_title\\t\"Cancer Genome Anatomy Project (CGAP) Human SAGE\"\\n!Series_summary\\t\"This series represents the Cancer Genome Anatomy Project SAGE library collection.  Libraries contained herein were either produced through CGAP funding, or donated to CGAP.\"\\n!Series_summary\\t\"The Cancer Genome Anatomy Project (CGAP: http://cgap.nci.nih.gov) is an interdisciplinary program established and administered by the National Cancer Institute (NCI: http://www.nci.nih.gov) to generate the information and technological tools needed to decipher the molecular anatomy of the cancer cell.\"\\n!Series_summary\\t\"\"\\n!Series_summary\\t\"SAGE libraries are named according to the following convention:\"\\n!Series_summary\\t\"* SAGE_Organ_histology_code_unique identifier, e.g., SAGE_Colon_adenocarcinoma_CL_Caco2\"\\n!Series_summary\\t\"* Codes: B = bulk; CL = cell line; CS = short-term cell culture; MD = micro-dissected; AP = antibody purified.\"\\n!Series_overall_design\\t\"324 libraries\"\\n\\nSample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\\n{0: [\\'organ/tissue: bone marrow\\', \\'organ/tissue: brain\\', \\'organ/tissue: cerebellum\\', \\'organ/tissue: mammary gland\\', \\'organ/tissue: cartilage\\', \\'organ/tissue: colon\\', \\'organ/tissue: stem cell\\', \\'organ/tissue: other\\', \\'organ/tissue: eye\\', \\'organ/tissue: uncharacterized tissue\\', \\'organ/tissue: heart\\', \\'organ/tissue: kidney\\', \\'organ/tissue: white blood cells\\', \\'organ/tissue: liver\\', \\'organ/tissue: lung\\', \\'organ/tissue: lymph node\\', \\'organ/tissue: muscle\\', \\'organ/tissue: ovary\\', \\'organ/tissue: pancreas\\', \\'organ/tissue: bone\\', \\'organ/tissue: peritoneum\\', \\'organ/tissue: placenta\\', \\'organ/tissue: prostate\\', \\'organ/tissue: retina\\', \\'organ/tissue: skin\\', \\'organ/tissue: spinal cord\\', \\'organ/tissue: stomach\\', \\'organ/tissue: gastrointestinal tract\\', \\'organ/tissue: testis\\', \\'organ/tissue: thyroid\\', \\'organ/tissue: uterus\\', \\'organ/tissue: vascular\\'], 1: [\\'tissue preparation: flow-sorted\\', \\'tissue preparation: antibody purified\\', \\'tissue preparation: bulk\\', \\'tissue preparation: cell line\\', \\'tissue preparation: short term culture\\', \\'tissue preparation: microscope dissected\\'], 2: [\\'sex: unknown\\', \\'age: fetal\\', \\'age: 34\\', \\'age: 35 yrs\\', \\'age: 23\\', \\'age: 9\\', \\'age: 10 yrs\\', \\'age: 53 yrs\\', \\'age: 5 yrs\\', \\'age: 66 yrs\\', \\'age: 5\\', \\'age: 30 yrs\\', \\'age: 13 yrs\\', \\'age: 65 yrs\\', \\'age: 38 yrs\\', \\'age: 37\\', \\'age: 43 yrs old\\', \\'age: 51 yr\\', \\'age: 71 yrs\\', \\'age: 32 yrs\\', \\'age: 9 months\\', \\'age: 4 years\\', \\'age: 6.3 years\\', \\'age: 29 yrs\\', \\'age: 7\\', \\'age: 27\\', \\'age: 17\\', \\'age: 1.7 yrs\\', \\'age: 7 yrs\\', \\'age: 31\\', \\'age: 3\\', \\'age: Spontaneously aborted male/female Caucasian fetuses, ages 20-33 weeks\\', \\'age: 4 mo\\', \\'age: 51\\', \\'age: 52 yrs\\', \\'age: 80 yrs\\', \\'age: 78 yrs\\', \\'cell type: glioblastoma,  cerebral cortex\\', \\'age: 67 yrs\\', \\'age: unknown\\', \\'age: 65\\', \\'cell type: glioblatastoma infected with  mutant EGFR, cerebral cortex\\', \\'cell type: glioblatastoma infected with LacZ control, cerebral cortex\\', \\'cell type: glioblastoma, cerebral cortex\\', \\'age: 63\\', \\'cell type: medulloblastoma, cerebellum\\', \\'age: 6\\', \\'age: 4\\', \\'age: 3.9\\', \\'age: 19\\', \\'age: 9 yrs\\', \\'age: 6 yrs\\', \\'age: 2.1 yrs\\', \\'age: 3.1 yrs\\', \\'age: 22\\', \\'age: 12 yrs\\', \\'age: 28\\', \\'age: 82\\', \\'cell type: normal, cerebral cortex\\', \\'sex: male\\', \\'age: 50\\', \\'age: 42\\', \\'sex: female\\', \\'age: 29\\', \\'age: 47\\', \\'age: 56.5\\', \\'age: 52.8\\', \\'age: 44 yrs\\', \\'age: 32\\', \\'age: 41\\', \\'age: 55\\', \\'age: 67\\', \\'age: 46\\', \\'age: 11\\', \\'age: 43\\', \\'age: 35\\', \\'age: 58\\', \\'age: 44\\', \\'age: 52\\', \\'age: 53\\', \\'age: 35 years old\\', \\'age: 28 years old\\', \\'age: 68\\', \\'age: 45 years old\\', \\'cell type: adenocarcinoma\\', \\'cell type: normal epithelium\\', \\'age: Passage 20 after derivation\\', \\'age: p31\\', \\'age: p54\\', \\'age: p22\\', \\'age: passage 22\\', \\'age: passage 33\\', \\'age: passage 38\\', \\'age: passage 16\\', \\'age: p36\\', \\'age: passage 50\\', \\'age: 69\\', \\'age: N/A\\', \\'age: 60\\', \\'age: 56\\'], 3: [\\'cell type: Stem Proginitor Cells, CD34+, CD38+\\', \\'cell type: Acute myelogenous leukemia\\', \\'cell type: normal bone marrow\\', nan, \\'cell type: normal fetal astrocyte, GFAP positive, cerebral cortex\\', \\'sex: male\\', \\'sex: female\\', \\'cell type: astrocytoma, right parietal occipital lobe, cerebral cortex\\', \\'sex: unknown\\', \\'other information: 5 pooled Duke glioblastoma multiforme primary tumors.\\', \\'other information: Brain, glioblastoma multiforme cell line, H54, that has been stably transfected with epidermal growth factor receptor vIII mutant\\', \\'other information: Brain, glioblastoma multiforme cell line, H54, that has been stably transfected with the b-galactosidase gene\\', \\'other information: negative for genomic amplification of EFGR, c-myc, n-myc and gli-1\\', \\'other information: medulloblastoma grade IV\\', \\'cell type: medulloblastoma cell line\\', \\'cell type: medulloblastoma, cerebellum\\', \\'cell type: benign meningioma of the meningothelial subtype\\', \\'cell type: atypical meningioma\\', \\'cell type: Malignant Meningioma\\', \\'cell type: normal cerebellum\\', \\'other information: Brain, normal, pooled. TB-1017 derived from a 51 yo male and TB-1060 derived from a 17 yo male.\\', \\'cell type: arachnoid cells, blood vessels, macrophages\\', \\'cell type: normal pediatric cortex\\', \\'other information: Short tags are extracted from the long SAGE data\\', \\'cell type: normal thalamus\\', \\'cell type: breast  carcinoma myoepithelium\\', \\'cell type: Grade II, ER+, PR+, Her2- tumor\\', \\'cell type: Grade I, ER+, PR+, Her2- invasive ductal\\', \\'cell type: carcinoma\\', \\'cell type: breast invasive ductal carcinoma\\', \\'cell type: breast carcinoma\\', \\'cell type: ER neagtive breast cancer cell line\\', \\'cell type: ER negative breast cancer cell line\\', \\'cell type: breast carcinoma cell line\\', \\'cell type: breast  carcinoma epithelium\\', \\'cell type: Grade I, ER+, PR+, Her2- invasive ductal carcinoma\\', \\'cell type: MUC1 bead purified cells from MCF10DCIS xenografts\\', \\'cell type: ITGB6 bead purified cells from MCF10DCIS xenografts\\', \\'cell type: estrogen responsive cells purified from normal mammary epithelium\\', \\'cell type: spontenously immortalized mammary myoepithelial cell line from Li-Fraumeni patient\\', \\'other information: Colon, primary tumor\\', \\'cell type: adenocarcinoma\\', \\'mutations: Wild type p53; RER+\\', \\'mutations: RER- with mutant p53 (248arg-trp) mutant APC\\', \\'cell type: cell line derived from colorectal carcinoma\\', \\'other information: Normal colonic epithelium\\', \\'cell type: normal\\', \\'other information: Kidney, embryonic cell line 293, induced cells, expressing beta-catenin\\', \\'other information: Uninduced 293 cells\\', \\'cell type: normal bulk liver\\', \\'cell type: poorly differentiated adenocarcinoma\\', \\'cell type: poor differentiated adenocarcinoma with lymphoplasmatic infiltration\\', \\'cell type: well differentiated adenocarcinoma with lymphoplasmatic infiltration\\', \\'cell type: normal muscle biopsy of vastus lateralis\\', \\'other information: Needle muscle biopsies were obtained from the vastus lateralis muscles of 8 healthy  young men (21-24 yr) under standard conditions. All were screened fro good health and normal neuromuscular function by a panel of laboratory tests and examination by neuromuscular disease specialist. None were involved in weight lifting or extreme levels of aerobic exercise. They were asked to refrain from activities more strenuous than walking for 3 days before the biopsy. They stayed at the University of Rochester GCRC the night before the biopsy, where they received a standard meal. They received a standard breakfast about 90 min before the biopsy. The tissue was removed whitin a few minutes of administration of the local anesthetic (lidocaine). It was quickly frozen in liquid nitrogen, then stored at -70 deg C until RNA extraction.\\', \\'other information: Ovary, primary ovarian tumour, serous adenocarcinoma.\\', \\'other information: Ovary, primary tumor. Serous adenocarcinoma.\\', \\'other information: Ovary, primary ovarian tumour. Serous adenocarcinoma.\\', \\'other information: Ovary, ovarian cancer cell line A2780\\', \\'other information: Ovarian Clear cell carcinoma cell line ES-2, poorly differentiated\\', \\'cell type: cystadenoma immortalized with SV40\\', \\'other information: Ovary, surface epithelium line\\', \\'cell type: normal surface epithelium\\', \\'other information: A frozen tumor sample was obtained from a colon cancer patient. Histopathology confirmed that the tumor sample contained at least 70% of tumor cells.\\', \\'other information: ATCC Library constructed in the Kern laboratory\\', \\'cell type: normal ductal\\', \\'cell type: malignant peritoneal mesothelioma\\', \\'cell type: normal first trimester placenta\\', \\'mutations: Inactivating mutation in PTEN gene\\', \\'cell type: carcinoma, prostate glands and stroma\\', \\'other information: Prostate, cancer cell line. Induced with synthetic androgen (R1881).\\', \\'other information: LNCaP cells were maintained in RPMI 1640 (Life Technologies, Inc., Gaithersburg, MD) supplemented with 10% fetal bovine serum (FBS, Life Technologies, Inc., Gaithersburg, MD) and experiments were performed on cells between passages 20 and 30. For the studies of androgen regulation, charcoal/dextran stripped androgen-free FBS (cFBS, Gemini Bio-Products, Inc., Calabasas, CA) was used. LNCaP cells were cultured first in RPMI 1640 with 10% cFBS for 5 days and then stimulated with an non-metabolizable androgen analog R1881 (DUPONT, Boston, MA) 10-8 M for 24 hours (for generating LNCaP-T library). LNCaP cells identically treated but without R1881 treatment served as control (for generating LNCaP-C library). Cells were harvested at indicated time and polyA+ RNA was double-selected with Fast Track kit (Invitrogene) SAGE libraries were generated according to the procedure provided by Dr. Kinzler (Johns Hopkins University School of Medicine, Baltimore, MD).\\', \\'other information: LNCaP cells were maintained in RPMI 1640 supplemented with 10% fetal bovine serum  and experiments were performed on cells between passages 20 and 30. For the studies of androgen regulation, charcoal/dextran stripped androgen-free FBS  was used. LNCaP cells were cultured first in RPMI 1640 with 10% cFBS for 5 days and then stimulated with an non-metabolizable androgen analog R1881 (DUPONT, Boston, MA) 10-8 M for 24 hours (for generating LNCaP-T library). LNCaP cells identically treated but without R1881 treatment served as control (for generating LNCaP-C library). Cells were harvested at indicated time and polyA+ RNA was double-selected with Fast Track kit (Invitrogene) SAGE libraries were generated according to the procedure provided by Dr. Kinzler (Johns Hopkins University School of Medicine, Baltimore, MD).\\', \\'cell type: prostate carcinoma\\', \\'cell type: normal epithelium & stroma\\', \\'cell type: normal epithelium\\', \\'sex: male and female\\', \\'cell type: synovial fibroblasts from rheumatoid arthritis\\', \\'other information: Library made from Poly A RNA supplied by Clonetech catalog #6573-1, and LOT #0031205.\\', \\'cell type: gastric cancer\\', \\'cell type: A pool of normal gastric epithelial tissues obtained from the antrum\\', \\'cell type: normal gastric epithelium\\', \\'cell type: mixture of human cancer cell lines\\', \\'cell type: high-grade comedo DCIS endothelium\\', \\'cell type: endothelium\\'], 4: [nan, \\'other information: Normal human astrocyte cells selected from the whole brain of a 22.5 wks old donor grown in the supplied media and harvested at passage 5. Cat# CC-2565, Lot 17657.\\', \\'cell type: pilocytic astrocyoma, cerebellum\\', \\'cell type: astrocytoma grade II\\', \\'cell type: Anaplastic GradeII Astrocytoma\\', \\'cell type: anaplastic astrocytoma\\', \\'cell type: Anaplastic Astrocytoma Grade III\\', \\'cell type: anaplastic gradeIII, primary, brain\\', \\'cell type: Anaplastic GradeIII Astrocytoma\\', \\'cell type: anaplastic astrocytoma grade III,\\', \\'cell type: Anaplastic astrocytoma grade III\\', \\'cell type: ependymoma\\', \\'other information: Short tags are extracted from Long SAGE data\\', \\'cell type: pediatric glioblastoma, cerebral cortex\\', \\'cell type: glioblastoma, Lt. Temporal, cerebral cortex\\', \\'cell type: Glioblastoma\\', \\'mutations: unknown\\', \\'cell type: glioblastoma, cerebral cortex\\', \\'cell type: GBM\\', \\'cell type: medulloblastoma, cerebellum\\', \\'cell type: medulloblastoma\\', \\'cell type: medulloblastoma with neuroblastic differentiation\\', \\'cell type: desmoplastic medulloblastoma\\', \\'cell type: large cell medulloblastoma\\', \\'cell type: anaplastic medulloblastoma\\', \\'cell type: medulloblastoma,\\', \\'cell type: Medulloblastoma Cell Line\\', \\'other information: Medulloblatoma cell line treated with Otx2 siRNA.\\', \\'mutations: c-myc oncogene genomic amplification\\', \\'mutations: has pten mutation, negative on screens for 17p loss, negative for c-myc or n-myc amplification\\', \\'cell type: Medulloblastoma\\', \\'cell type: normal cerebellum\\', \\'cell type: normal whitematter, cerebral cortex\\', \\'other information: autopsy leptomeninges\\', \\'other information: Normal frontal cortex, bulk tissue, from 15 month male seizure patient (partial lobectomy).\\', \\'cell type: oligodendroglioma\\', \\'cell type: well-differentiated oligodendroglioma, frontal lobe, cortex\\', \\'other information: Extensive high-grade comedo DCIS (~8-10cm). Purified using anti-CD10 antibody and magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-MYOEP-6 myoepithelial cells\"\\', \\'cell type: breast  carcinoma myoepithelium\\', \\'cell type: ER+, PR+, HER2-, grade II\\', \\'other information: CD10+ purified myofibroblasts from invasive breast cancer. Library name in Cancer Cell, 2004, 6(1):17-32 \"I-MYOFIB-8 myofibroblasts\"\\', \\'other information: CD10+ purified myofibroblasts from invasive breast cancer  Library name in Cancer Cell, 2004, 6(1):17-32 \"I-MYOFIB-9-myofibroblasts\"\\', \\'cell type: ductal carcinoma in situ, extensive, gradeIII, Her2+\\', \\'cell type: invasive breast cancer ER+, PR+, Her2-, grade II\\', \\'cell type: carcinoma\\', \\'cell type: primary ductal in situ breast carcinoma, intermediate grade\\', \\'cell type: primary grade III. ductal invasive breast carcinoma\\', \\'other information: Ductal carcinoma in situ, high-grade, comedo.\\', \\'other information: Invasive ductal carcinoma with some in situ component, low-grade, ER-positive, erbB22-negative, lymph node-negative, p53-negative.\\', \\'other information: Breast, invasive ductal carcinoma, high-grade, ERneg, erbB2neg., lymph node neg., p53pos.\\', \\'other information: Breast, invasive ductal carcinoma, low-grade, ERpos, erbB2neg., lymph node neg., p53neg.\\', \\'other information: Estrogen receptor positive MCF7 cells were grown in conditions devoid of estrogen for 48h (i.e. IMEM media supplemented with 10% Charcoal Stripped Serum). This SAGE library represents the 0h time point (i.e. baseline, untreated cells) of an estrogen exposure time course experiment.\\', \\'other information: cells were collected 24 hours after expression of HIN-1\\', \\'other information: cells were collected 48 hours after expression of HIN-1\\', \\'other information: ZR75-1 estrogen receptor positive, estrogen dependent human breast cancer cell line, estrogen deprived for a week, cells were collected 16 hours following addition of fresh medium containing 10 nM estrogen.\\', \\'other information: ZR75-1 estrogen receptor positive, estrogen dependent human breast cancer cell line, estrogen deprived for a week, cells were collected 16 hours following addition of fresh medium containing 10 uM 4-OH-tamoxifen.\\', \\'other information: ZR75-1 human estrogen receptor positive, estrogen dependent breast cancer cell line, estrogen deprived for 1 week, cells collected 16 hours following change to fresh medium (no estrogen added).\\', \\'other information: Breast, DCIS, BerEp4 purified cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-2 epithelial cells\"\\', \\'other information: Extensive high-grade comedo DCIS (~8-10 cm). Purified using BerEP4 beads (Dynal).Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-6 epithelial cells\"\\', \\'cell type: breast  carcinoma epithelium\\', \\'other information: BerEP4 purified epithelial cells from invasive breast cancer. Library name in Cancer Cell, 2004, 6(1):17-32 \"I-EPI-8 epithelial cells\"\\', \\'other information: BerEP4 purified epithelial cells from invasive breast cancer. Library name in Cancer Cell, 2004, 6(1):17-32 \"I-EPI-9 epithelial cells\"\\', \\'cell type: invasive breast cancer, ER+, PR+, Her-, grade II\\', \\'cell type: Invasive ductal carcinoma, ER+,PR+,HER2-\\', \\'cell type: Invasive ductal carcinoma, high-grade, ER+,PR+,HER2-\\', \\'cell type: Pleural effusion, HER2-, recurrence of ER+ tumor\\', \\'cell type: Ascites, mixed lobular/ductal carcinoma, HER2-, recurrence of ER+ tumor\\', \\'cell type: invasive ductal carcinoma ER+,PR+,HER2-\\', \\'mutations: 8q gain, CDKN1A and LRP1B homozygous deletions\\', \\'cell type: ductal carcinoma in situ\\', \\'cell type: extensive LCIS\\', \\'cell type: breast carcinoma metastasis to lung\\', \\'cell type: juvenile fibroadenoma\\', \\'cell type: metastasis from breast tumor\\', \\'cell type: breast hyperplasia\\', \\'cell type: normal epithelium\\', \\'cell type: normal\\', \\'cell type: Normal breast tissue reduction\\', \\'cell type: normal breast tissue reduction\\', \\'other information: cells purified using MUC1 antibody, these cells are MUC1+ mammary epithelial cells\\', \\'cell type: normal breast tissue from a breast cancer patient (corresponding to IDC7)\\', \\'cell type: breast myoepithelium\\', \\'mutations: Li-Fraumeni patient (heterozygous TP53 mutation)\\', \\'cell type: normal breast organoid\\', \\'cell type: normal breast stroma\\', \\'cell type: recurrent phyllodes tumor, malignant, high grade, poorly differentiated\\', \\'cell type: chondrosarcoma grade 2.\\', \\'cell type: Dedifferentiated chondrosarcoma lung metastasis\\', \\'mutations: RER- (replication error negitive cell line), has normal mismatch repair\\', \\'mutations: Ras gene codon 13; wild type p53; RER+\\', \\'other information: Colon, cancer cell line.\\', \\'other information: Colon, cancer cell line. Mismatch proficient(RER-) with a mutant p53(248arg -> trp) and a mutant APC.\\', \\'cell type: adenocarcinoma\\', \\'cell type: high grade dysplasia\\', \\'cell type: low grade dysplasia\\', \\'mutations: N/A\\', \\'other information: SV-40 treated fibroblasts grown in culture after crisis in which the ALT pathway of telomerase maintenance is used\\', \\'cell type: Adenocarcinoma\\', \\'cell type: Adenocarcinoma, tubular, poorly differentiated\\', \\'cell type: Adenocarcinoma, tubular\\'], 5: [nan, \\'other information: grade one(1) pilocytic astrocytoma\\', \\'other information: Malignant embryonal tumor consistent with ependymoblastoma, in patient who subsequently had recurrent ependymal tumors\\', \\'other information: recurrent glioma from patient with previous ependymoblastoma in same fronto-parietal region\\', \\'other information: recurrent Neuroectodermal tumor w/epend diff. tumor, from 6.3 year old female with history of ependymoma\\', \\'other information: primary ependymoma in 34 yo female\\', \\'other information: primary brain tumor in right foramen of lushka\\', \\'other information: Grade II\\', \\'other information: Suboccipital Craniectomy\\', \\'other information: Lt. parietal occipital infantile CNS tumor consistent with glioblastoma multiforme\\', \\'mutations: No 17p loss or p53 mutations detected, Rasheed et al Can. Res. 54 p1324\\', \\'other information: Brain, c-myc positive medulloblastoma cell line H341\\', \\'other information: Brain, cell line mhh-1\\', \\'other information: Brain, normal, greater than 95% white matter. Post-mortem delay of 3:15. 82 year old male.\\', \\'other information: Tumor grade 2/4\\', \\'other information: Brain, well differentiated oligodendroglioma\\', \\'other information: Invasive breast tumor (~2cm), extensive high-grade comedo DCIS (~4 cm), positive lymph nodes. Purified using anti-CD10 antibody and magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-MYOEP-7 myoepithelial cells\"\\', \\'mutations: ER-, PR-\\', \\'mutations: ER+, PR+\\', \\'other information: 1. Short SAGE tags are extracted from Long SAGE data.    2. Estrogen and progesteron receptor positive, erbB2 negative, 0/25 lymph nodes (no lymph node metastasis)\\', \\'other information: Short tags are extracted from the long SAGE data\\', \\'other information: Ductal carcinoma in situ, high-grade, comedo from 32 yo woman.\\', \\'other information: Breast, ductal carcinoma in situ, high-grade, comedo, Ber-EP4 purified cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-3 epithelial cells\"\\', \\'other information: Invasive breast tumor(~2cm), extensive high-grade comedo DCIS (~4 cm), positive lymph nodes. Purified using BerEP4 beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-EPI-7 epithelial cells\"\\', \\'other information: BerEP4 purified epithelial cells\\', \\'other information: cells purified using CD24 antibody, these cells are CD24+ more differentiated breast cancer cells\\', \\'other information: cells purified using CD44 antibody, these cells are CD44+ breast \"cancer stem\" cells\\', \\'other information: cells purified using PROCR antibody, these cells are CD44+ breast \"cancer stem\" cells\\', \\'other information: Breast, macroscopically dissected (>90% pure)  high grade, comedo DCIS (ductal carcinoma in situ) of the breast obtained from a 67 year old female patient undergoing partial mastectomy. The tumor cells are estrogen and progesteron receptor positive\\', \\'mutations: Grade III, ERpos, PRlow, Her2neg\\', \\'other information: Library name in Cancer Cell, 2004, 6(1):17-32 \"FA\"\\', \\'other information: Metastatic tumor (lymph node metastasis from breast primary tumor). Age at discovery of primary tumor: 56.5 years, 10 postive lymph nodes, Primary tumor is 3 cm, histo grade = 3, nuclear grade = 3, ER-, PR-.\\', \\'other information: Metastatic tumor (lymph node metastasis from breast primary tumor). Age at discovery of primary tumor: 52.8 years, 16 postive lymph nodes.  Primary tumor is 3.2 cm, histo grade = 3, nuclear grade = 3, ER+, PR+.\\', \\'other information: normal mammary epithelium (gestational hyperplasia) from a 31 year old breast cancer patient (20 weeks pregnant).\\', \\'other information: Normal human luminal mammary epithelial cells purified with Epithelial Enrich (BER-EP4 antibody conjugated to magnetic beads) tissue obtained from a 43 year old female patient undergoing reduction mammoplasty. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-EPI-1 epithelial cells\"\\', \\'other information: Breast, normal luminal mammary epithelium, BerEp4 purified cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"n-EPI-2 epithelial cells\"\\', \\'other information: cells purified using CD24 antibody, these cells are CD24+ mammary epithelial cells\\', \\'other information: cells purified using CD24 antibody, these cells are CD24+ differentiated luminal epithelial cells\\', \\'other information: cells purified using CD44 antibody, these cells are CD44+ mammary epithelial stem cells\\', \\'other information: cells purified using MUC1 antibody, these cells are MUC1+ mammary epithelial cells\\', \\'other information: CD10 positive purified myoepithelial cells. Short tags extracted from long. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-MYO-I7 myoepithelial cells\"\\', \\'other information: normal mammary myoepithelial cells from breast reduction tissue of a 27 year old female. Antibody purified with CD10 to obtain myoepithelial cells. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-MYOEP-1 myoepithelial cells\"\\', \\'other information: cells purified using CD10 antibody, these cells are CD10+ mammary myoepithelial cells\\', \\'other information: Normal breast organoids from reduction mammoplasty.\\', \\'other information: Normal breast tissue from reduction mammoplasty. Library name in Cancer Cell, 2004, 6(1):17-32 \"N-STR-1-fibroblasts\"\\', \\'other information: Library name in Cancer Cell, 2004, 6(1):17-32 \"PHY\"\\', \\'mutations: N/A\\', \\'other information: Colon, colorectal carcinoma cell line (RER-), ATCC: HTB-37.\\', \\'other information: cell line derived from colorectal carcinoma. ATCC: CCL-247. Mutation in the Ras gene, codon 13. Wild type p53. RER+\\', \\'other information: Total RNA from Ambion, Cat #7966, Lot #051P43C\\', \\'other information: Source:  Clontech Total RNA, Cat.# 64012-1, Lot# 2120258.  Right kidney; poorly differentiated with squamous differentiation (grade III-IV); extensive invasion seen in kidney, including renal tubules, lymphovascular areas and perinephric fat; no evidence of tumor in ureter, artery or vein.\\', \\'other information: Normal kidney female. Cause of death IC brain hemorrhage. Poly A RNA from Ambion Cat#7977 lot #100P94B\\', \\'other information: Pooled Femal Adult total RNA from Stratagene, cat#:735016, lot#0110204\\', \\'other information: Ambion cat #7968, lot #011P27A. Whole lung RNA from motor vehicle accident victim.\\', \\'other information: Full-term placenta.\\', \\'other information: PC3 cells, AS2 treatment, total number of 42,236 SAGE tags sequenced. 40,768 SAGE tags analyzed after clean-up (extraction of poly(A) tags, linker-generated tags, and tags originated from clones producing 4 tags or less).\\', \\'other information: Libraries were constructed using RNA isolated from pooled 4 mm diameter punches of the macula (encompassing the fovea) and matched 4 mm diameter punches of pooled mid-peripheral retina from the same five donors designated, \"4Mac\" and \"4Peri\", respectively.\\', \\'other information: The library was constructed using RNA isolated from pooled 4 mm punches of human macula and central peripheral retina from 7 morphologically normal donor eyes.\\', \\'other information: Patient had periocular malignancies that required exenteration of a normal eye as part of the clinically indicated surgical therapy. The eye was placed in saline after it was removed from the patient and peripheral retinal fragments were collected from regions outside the central arcades and frozen within 45 minutes of the surgical removal to ensure minimal degradation of mRNA molecules.\\', \\'other information: A normal peripheral retina from a 44-year-old male. Patient had periocular malignancies that required exenteration of a normal eye as part of the clinically indicated surgical therapy. The eye was placed in saline after it was removed from the patient and peripheral retinal fragments were collected from regions outside the central arcades and frozen within 45 minutes of the surgical removal to ensure minimal degradation of mRNA molecules.\\', \\'other information: RPE/choroid libraries were made from 4 mm diameter punches of macular RPE/choroid (4MacRPE) and matching 4 mm diameter punches of mid-peripheral RPE/choroid, (4PeriRPE), taken from five donor eyes.  While these libraries were made from RPE/choroid, subsequent analysis has revealed significant content of retinal transcripts in addition to RPE markers.\\', \\'other information: Gastroesophageal junction adenocarcinoma, T4N0M0, moderately differentiated, obtained by surgery.\\', \\'other information: lower esophagus close to gastroesophageal junction adenocarcinoma, T4N0M0, moderately differentiated, obtained by surgery.\\', \\'other information: Gastroesophageal junction adenocarcinoma, T4N0M0, moderately differentiated, obtained by surgery. Xenograft.\\', \\'other information: Normal breast tissue from reduction mammoplasty. Purified using anti-P1H12 antibody and magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"N-END-1-endothelial cells\"\\', \\'other information: White blood cells, isolated from a mammary gland ductal carcinoma. Purified using CD45+CD15 coupled  magnetic beads (Dynal). Extensive high-grade comedo DCIS (~8-10cm). Library name in Cancer Cell, 2004, 6(1):17-32 \"D-LEU-6-leukocytes\"\\', \\'other information: Library name in Cancer Cell, 2004, 6(1):17-32 \"D-LEU-7-leukocytes\"\\', \\'other information: CD45+ purified tumor  infiltrating leukocytes\\', \\'mutations: none\\', \\'other information: White blood cells isolated from normal breast tissue from reduction mammoplasty. Purified using CD45+CD15 coupled magnetic beads (Dynal). Library name in Cancer Cell, 2004, 6(1):17-32 \"N-LEU-1-leukocytes\"\\'], 6: [nan, \\'other information: Invasive ductal carcinoma - primary tumor.  Age at discovery 56.5 years, 10 postive lymph nodes, 3 cm primary tumor, histo grade = 3, nuclear grade = 3, ER-, PR-.\\', \\'other information: Invasive ductal carcinoma - primary tumor.  Age at discovery 52.8 years, 16 postive lymph nodes, 3.2 cm primary tumor, histo grade = 3, nuclear grade = 3, ER+, PR+.\\', \\'other information: Lung metastasis of a breast cancer.\\', \\'other information: Tissue section of left pelvis from 35 years old male with grade 2 recurrent chondrosarcoma was used for mRNA isolation.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of nterior medial aspect of the pelvis from 28 years old male with grade 2 recurrent chondrosarcoma was divided, by gross analysis, into a cartilage-like sample and was used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in growth medium that contains heat inactivated FBS (Life Technologies Gibco) (12%), DMEM, glucose (4.5 gm/l), HEPES buffer (25 mM) solution and gentamicin sulfate (50 â▌g/ml) (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of nterior medial aspect of the pelvis from 28 years old male with grade 2 recurrent chondrosarcoma was divided, by gross analysis, into a fibrosis-like sample and was used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in growth medium that contains heat inactivated FBS (Life Technologies Gibco) (12%), DMEM, glucose (4.5 gm/l), HEPES buffer (25 mM) solution and gentamicin sulfate (50 â▌g/ml) (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of lung right lower lobe, middle posterior metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of lung right middle lobe, upper anterior   metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of lung right lower lobe, middle diaphragmatic metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of lung right upper lobe, apical lateral metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\', \\'other information: Tissue section of lung left lower lobe, diaphragmatic anterior metastatic lesion from 45 years old female with metastatic dedifferentiated chondrosarcoma were used to obtain mono-dispersed cells. The cell line was established from the cells after culturing in RPMI-1640 supplemented with 10% FBS and 0.1% gentamicin sulfate (Gemini Bioproducts). The cell line had gone through 4 doublings to yield 2 x 107 cells.  The cell line was routinely screened for mycoplasma sps. Using a PCR-based ELISA detection assay (Roche). Cell manipulations were always performed on 80-90% confluent cultures for consistency.  mRNA was isolated using a direct mRNA extraction kit.\\']}\\n'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":149}],"source":["f'''As a biomedical research team, we are selecting datasets to study the association between the human trait \\'{TRAIT}\\' and genetic factors, optionally considering the influence of age and gender. After searching the GEO database and parsing the matrix file of a series, we obtained background information and sample characteristics data. We will provide textual information about the dataset background, and a Python dictionary storing a list of unique values for each field of the sample characteristics data. Please carefully review the provided information and answer the following questions about this dataset:\n","1. Does this dataset contain gene expression data? (Note: Pure miRNA data is not suitable.)\n","2. For each of the traits \\'{TRAIT}\\', 'age', and 'gender', please address these points:\n","   (1) Is there human data available for this trait?\n","   (2) If so, identify the key in the sample characteristics dictionary where unique values of this trait is recorded. The key is an integer. The trait information might be explicitly recorded, or can be inferred from the field with some biomedical knowledge or understanding about the data collection process.\n","   (3) Choose an appropriate data type (either 'continuous' or 'binary') for each trait. Write a Python function to convert any given value of the trait to this data type. The function should handle inference about the trait value and convert unknown values to None.\n","   Name the functions 'convert_trait', 'convert_age', and 'convert_gender', respectively.\n","\n","Background information about the dataset:\n","{background_info}\n","\n","Sample characteristics dictionary (from \"!Sample_characteristics_ch1\", converted to a Python dictionary that stores the unique values for each field):\n","{clinical_data_unique}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508232400Z","start_time":"2023-12-31T21:19:35.492903487Z"},"colab":{"base_uri":"https://localhost:8080/","height":178},"id":"VO8VaFm6ZWQa","executionInfo":{"status":"ok","timestamp":1704462323565,"user_tz":-480,"elapsed":519,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"0b03913a-ceb7-4eef-c844-9d3a5b24b9fb"}},{"cell_type":"markdown","source":["Understand and verify the answer from GPT, to assign values to the below variables. Assign None to the 'row_id' variables if relevant data row was not found.\n","Later we need to let GPT format its answer to automatically do these. But given the complexity of this step, let's grow some insight from the free-text answers for now."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978240504Z","start_time":"2023-12-31T03:58:04.922365324Z"},"id":"00OpftSgZWQb"}},{"cell_type":"code","execution_count":150,"outputs":[],"source":["is_gene_availabe = False\n","trait_row_id = 3\n","age_row_id = 2\n","gender_row_id = 2\n","\n","trait_type = 'binary'"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508295660Z","start_time":"2023-12-31T21:19:35.493015671Z"},"id":"boW_XbcAZWQb","executionInfo":{"status":"ok","timestamp":1704462404357,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":151,"outputs":[{"output_type":"stream","name":"stdout","text":["This cohort is not usable. Please skip the following steps and jump to the next accession number.\n"]}],"source":["is_available = is_gene_availabe and (trait_row_id is not None)\n","if not is_available:\n","    save_cohort_info(cohort, JSON_PATH, is_available)\n","    print(\"This cohort is not usable. Please skip the following steps and jump to the next accession number.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508365153Z","start_time":"2023-12-31T21:19:35.493151450Z"},"id":"oHwwe4oPZWQb","executionInfo":{"status":"ok","timestamp":1704462406458,"user_tz":-480,"elapsed":3,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"5f57dbf4-2d00-43ce-c1d1-d5f76a6494a5"}},{"cell_type":"code","execution_count":116,"outputs":[],"source":["# Verify and use the functions generated by GPT\n","\n","def convert_trait(value):\n","    if 'glioma grade' in value:\n","        return 1\n","    else:\n","        return None\n","\n","\n","def convert_age(age_string):\n","    \"\"\"\n","    Convert age string to a numerical value.\n","    \"\"\"\n","    if age_string.startswith('age: '):\n","        try:\n","            return int(age_string.split('age: ')[1].rstrip('y'))  # Extracting numeric part\n","        except ValueError:\n","            return None  # In case of parsing error\n","    else:\n","        return None  # If age string does not follow expected format\n","\n","\n","\n","def convert_gender(value):\n","    # For 'gender': Binary\n","    if 'female' in value.lower():\n","        return '1'\n","    elif 'male' in value.lower():\n","        return '0'\n","    else:\n","        return None  # Unknown gender\n","\n"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508426069Z","start_time":"2023-12-31T21:19:35.493220101Z"},"id":"uFVuqxemZWQb","executionInfo":{"status":"ok","timestamp":1704461886823,"user_tz":-480,"elapsed":487,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":117,"outputs":[{"output_type":"execute_result","data":{"text/plain":["                   GSM590744 GSM590745 GSM590746 GSM590747 GSM590748  \\\n","Lower Grade Glioma         1         1         1         1         1   \n","Age                       72        70        34        54        34   \n","Gender                     0         1         0         0         0   \n","\n","                   GSM591700 GSM591701 GSM591702 GSM591703 GSM591704  ...  \\\n","Lower Grade Glioma         1         1         1         1         1  ...   \n","Age                       68        30        60        73        52  ...   \n","Gender                     0         1         0         0         1  ...   \n","\n","                   GSM591812 GSM591815 GSM591817 GSM591818 GSM591819  \\\n","Lower Grade Glioma         1         1         1         1         1   \n","Age                       69        36        38        63        74   \n","Gender                     1         1         1         0         0   \n","\n","                   GSM591829 GSM591830 GSM591831 GSM591832 GSM591833  \n","Lower Grade Glioma         1         1         1         1         1  \n","Age                       67        54        46        55        72  \n","Gender                     1         0         0         1         0  \n","\n","[3 rows x 32 columns]"],"text/html":["\n","  <div id=\"df-21120246-fb5c-4cd2-b8ae-6684ee872afa\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>GSM590744</th>\n","      <th>GSM590745</th>\n","      <th>GSM590746</th>\n","      <th>GSM590747</th>\n","      <th>GSM590748</th>\n","      <th>GSM591700</th>\n","      <th>GSM591701</th>\n","      <th>GSM591702</th>\n","      <th>GSM591703</th>\n","      <th>GSM591704</th>\n","      <th>...</th>\n","      <th>GSM591812</th>\n","      <th>GSM591815</th>\n","      <th>GSM591817</th>\n","      <th>GSM591818</th>\n","      <th>GSM591819</th>\n","      <th>GSM591829</th>\n","      <th>GSM591830</th>\n","      <th>GSM591831</th>\n","      <th>GSM591832</th>\n","      <th>GSM591833</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>Lower Grade Glioma</th>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>...</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>Age</th>\n","      <td>72</td>\n","      <td>70</td>\n","      <td>34</td>\n","      <td>54</td>\n","      <td>34</td>\n","      <td>68</td>\n","      <td>30</td>\n","      <td>60</td>\n","      <td>73</td>\n","      <td>52</td>\n","      <td>...</td>\n","      <td>69</td>\n","      <td>36</td>\n","      <td>38</td>\n","      <td>63</td>\n","      <td>74</td>\n","      <td>67</td>\n","      <td>54</td>\n","      <td>46</td>\n","      <td>55</td>\n","      <td>72</td>\n","    </tr>\n","    <tr>\n","      <th>Gender</th>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>...</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>3 rows × 32 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-21120246-fb5c-4cd2-b8ae-6684ee872afa')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-21120246-fb5c-4cd2-b8ae-6684ee872afa button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-21120246-fb5c-4cd2-b8ae-6684ee872afa');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-a9e8e541-a862-49c1-b831-08aee655ecb7\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-a9e8e541-a862-49c1-b831-08aee655ecb7')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-a9e8e541-a862-49c1-b831-08aee655ecb7 button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":117}],"source":["selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row_id, age_row_id, gender_row_id,\n","                                                      convert_trait, convert_age, convert_gender)\n","selected_clinical_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.508533013Z","start_time":"2023-12-31T21:19:35.493272992Z"},"colab":{"base_uri":"https://localhost:8080/","height":228},"id":"3kQDZKKHZWQb","executionInfo":{"status":"ok","timestamp":1704461889077,"user_tz":-480,"elapsed":514,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"f131e32b-befa-44b0-baf0-d4c2432f69c6"}},{"cell_type":"markdown","source":["### Genetic data preprocessing and final filtering"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978664193Z","start_time":"2023-12-31T03:58:04.966117261Z"},"id":"FYVzuozPZWQb"}},{"cell_type":"code","execution_count":118,"outputs":[{"output_type":"execute_result","data":{"text/plain":["          ID  GSM590744  GSM590745  GSM590746  GSM590747  GSM590748  \\\n","0  1007_s_at   3.642191   3.701979   3.549620   3.668898   3.646627   \n","1    1053_at   3.024226   3.072491   3.178727   2.938471   2.975731   \n","2     117_at   3.301488   2.976024   3.064205   2.390648   3.175775   \n","3     121_at   3.326044   3.210726   3.272794   3.345455   3.477646   \n","4  1255_g_at   2.699437   2.523530   2.769053   2.914576   2.695694   \n","\n","   GSM591700  GSM591701  GSM591702  GSM591703  ...  GSM591812  GSM591815  \\\n","0   3.717462   3.682097   3.667317   3.641540  ...   3.662805   3.677109   \n","1   3.053342   3.074540   3.100501   3.124698  ...   3.169326   3.138912   \n","2   3.036856   2.890154   2.939648   2.939885  ...   2.980282   2.941577   \n","3   3.229819   3.343052   3.398781   3.324753  ...   3.315291   3.318474   \n","4   2.660008   2.538062   2.724338   2.828592  ...   2.640814   3.087213   \n","\n","   GSM591817  GSM591818  GSM591819  GSM591829  GSM591830  GSM591831  \\\n","0   3.647491   3.583830   3.605021   3.490284   3.574239   3.557329   \n","1   3.034664   3.027387   3.170743   3.152728   2.986613   2.884332   \n","2   2.948396   3.127264   2.842070   2.998904   2.879601   2.918557   \n","3   3.438588   3.484291   3.390079   3.380583   3.377112   3.342801   \n","4   2.715261   2.945468   2.778569   2.793341   2.852065   2.653127   \n","\n","   GSM591832  GSM591833  \n","0   3.643160   3.618054  \n","1   3.196361   3.169115  \n","2   2.970701   2.860472  \n","3   3.411945   3.367096  \n","4   2.894916   2.495179  \n","\n","[5 rows x 33 columns]"],"text/html":["\n","  <div id=\"df-e16e23fa-9362-4080-a562-f3e05b7ab618\" class=\"colab-df-container\">\n","    <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ID</th>\n","      <th>GSM590744</th>\n","      <th>GSM590745</th>\n","      <th>GSM590746</th>\n","      <th>GSM590747</th>\n","      <th>GSM590748</th>\n","      <th>GSM591700</th>\n","      <th>GSM591701</th>\n","      <th>GSM591702</th>\n","      <th>GSM591703</th>\n","      <th>...</th>\n","      <th>GSM591812</th>\n","      <th>GSM591815</th>\n","      <th>GSM591817</th>\n","      <th>GSM591818</th>\n","      <th>GSM591819</th>\n","      <th>GSM591829</th>\n","      <th>GSM591830</th>\n","      <th>GSM591831</th>\n","      <th>GSM591832</th>\n","      <th>GSM591833</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>1007_s_at</td>\n","      <td>3.642191</td>\n","      <td>3.701979</td>\n","      <td>3.549620</td>\n","      <td>3.668898</td>\n","      <td>3.646627</td>\n","      <td>3.717462</td>\n","      <td>3.682097</td>\n","      <td>3.667317</td>\n","      <td>3.641540</td>\n","      <td>...</td>\n","      <td>3.662805</td>\n","      <td>3.677109</td>\n","      <td>3.647491</td>\n","      <td>3.583830</td>\n","      <td>3.605021</td>\n","      <td>3.490284</td>\n","      <td>3.574239</td>\n","      <td>3.557329</td>\n","      <td>3.643160</td>\n","      <td>3.618054</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1053_at</td>\n","      <td>3.024226</td>\n","      <td>3.072491</td>\n","      <td>3.178727</td>\n","      <td>2.938471</td>\n","      <td>2.975731</td>\n","      <td>3.053342</td>\n","      <td>3.074540</td>\n","      <td>3.100501</td>\n","      <td>3.124698</td>\n","      <td>...</td>\n","      <td>3.169326</td>\n","      <td>3.138912</td>\n","      <td>3.034664</td>\n","      <td>3.027387</td>\n","      <td>3.170743</td>\n","      <td>3.152728</td>\n","      <td>2.986613</td>\n","      <td>2.884332</td>\n","      <td>3.196361</td>\n","      <td>3.169115</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>117_at</td>\n","      <td>3.301488</td>\n","      <td>2.976024</td>\n","      <td>3.064205</td>\n","      <td>2.390648</td>\n","      <td>3.175775</td>\n","      <td>3.036856</td>\n","      <td>2.890154</td>\n","      <td>2.939648</td>\n","      <td>2.939885</td>\n","      <td>...</td>\n","      <td>2.980282</td>\n","      <td>2.941577</td>\n","      <td>2.948396</td>\n","      <td>3.127264</td>\n","      <td>2.842070</td>\n","      <td>2.998904</td>\n","      <td>2.879601</td>\n","      <td>2.918557</td>\n","      <td>2.970701</td>\n","      <td>2.860472</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>121_at</td>\n","      <td>3.326044</td>\n","      <td>3.210726</td>\n","      <td>3.272794</td>\n","      <td>3.345455</td>\n","      <td>3.477646</td>\n","      <td>3.229819</td>\n","      <td>3.343052</td>\n","      <td>3.398781</td>\n","      <td>3.324753</td>\n","      <td>...</td>\n","      <td>3.315291</td>\n","      <td>3.318474</td>\n","      <td>3.438588</td>\n","      <td>3.484291</td>\n","      <td>3.390079</td>\n","      <td>3.380583</td>\n","      <td>3.377112</td>\n","      <td>3.342801</td>\n","      <td>3.411945</td>\n","      <td>3.367096</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>1255_g_at</td>\n","      <td>2.699437</td>\n","      <td>2.523530</td>\n","      <td>2.769053</td>\n","      <td>2.914576</td>\n","      <td>2.695694</td>\n","      <td>2.660008</td>\n","      <td>2.538062</td>\n","      <td>2.724338</td>\n","      <td>2.828592</td>\n","      <td>...</td>\n","      <td>2.640814</td>\n","      <td>3.087213</td>\n","      <td>2.715261</td>\n","      <td>2.945468</td>\n","      <td>2.778569</td>\n","      <td>2.793341</td>\n","      <td>2.852065</td>\n","      <td>2.653127</td>\n","      <td>2.894916</td>\n","      <td>2.495179</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 33 columns</p>\n","</div>\n","    <div class=\"colab-df-buttons\">\n","\n","  <div class=\"colab-df-container\">\n","    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e16e23fa-9362-4080-a562-f3e05b7ab618')\"\n","            title=\"Convert this dataframe to an interactive table.\"\n","            style=\"display:none;\">\n","\n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n","    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n","  </svg>\n","    </button>\n","\n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    .colab-df-buttons div {\n","      margin-bottom: 4px;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","    <script>\n","      const buttonEl =\n","        document.querySelector('#df-e16e23fa-9362-4080-a562-f3e05b7ab618 button.colab-df-convert');\n","      buttonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","      async function convertToInteractive(key) {\n","        const element = document.querySelector('#df-e16e23fa-9362-4080-a562-f3e05b7ab618');\n","        const dataTable =\n","          await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                    [key], {});\n","        if (!dataTable) return;\n","\n","        const docLinkHtml = 'Like what you see? Visit the ' +\n","          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","          + ' to learn more about interactive tables.';\n","        element.innerHTML = '';\n","        dataTable['output_type'] = 'display_data';\n","        await google.colab.output.renderOutput(dataTable, element);\n","        const docLink = document.createElement('div');\n","        docLink.innerHTML = docLinkHtml;\n","        element.appendChild(docLink);\n","      }\n","    </script>\n","  </div>\n","\n","\n","<div id=\"df-054463a8-10f5-4de4-aa23-b85ce1864d6b\">\n","  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-054463a8-10f5-4de4-aa23-b85ce1864d6b')\"\n","            title=\"Suggest charts\"\n","            style=\"display:none;\">\n","\n","<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","     width=\"24px\">\n","    <g>\n","        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n","    </g>\n","</svg>\n","  </button>\n","\n","<style>\n","  .colab-df-quickchart {\n","      --bg-color: #E8F0FE;\n","      --fill-color: #1967D2;\n","      --hover-bg-color: #E2EBFA;\n","      --hover-fill-color: #174EA6;\n","      --disabled-fill-color: #AAA;\n","      --disabled-bg-color: #DDD;\n","  }\n","\n","  [theme=dark] .colab-df-quickchart {\n","      --bg-color: #3B4455;\n","      --fill-color: #D2E3FC;\n","      --hover-bg-color: #434B5C;\n","      --hover-fill-color: #FFFFFF;\n","      --disabled-bg-color: #3B4455;\n","      --disabled-fill-color: #666;\n","  }\n","\n","  .colab-df-quickchart {\n","    background-color: var(--bg-color);\n","    border: none;\n","    border-radius: 50%;\n","    cursor: pointer;\n","    display: none;\n","    fill: var(--fill-color);\n","    height: 32px;\n","    padding: 0;\n","    width: 32px;\n","  }\n","\n","  .colab-df-quickchart:hover {\n","    background-color: var(--hover-bg-color);\n","    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n","    fill: var(--button-hover-fill-color);\n","  }\n","\n","  .colab-df-quickchart-complete:disabled,\n","  .colab-df-quickchart-complete:disabled:hover {\n","    background-color: var(--disabled-bg-color);\n","    fill: var(--disabled-fill-color);\n","    box-shadow: none;\n","  }\n","\n","  .colab-df-spinner {\n","    border: 2px solid var(--fill-color);\n","    border-color: transparent;\n","    border-bottom-color: var(--fill-color);\n","    animation:\n","      spin 1s steps(1) infinite;\n","  }\n","\n","  @keyframes spin {\n","    0% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","      border-left-color: var(--fill-color);\n","    }\n","    20% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    30% {\n","      border-color: transparent;\n","      border-left-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","      border-right-color: var(--fill-color);\n","    }\n","    40% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-top-color: var(--fill-color);\n","    }\n","    60% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","    }\n","    80% {\n","      border-color: transparent;\n","      border-right-color: var(--fill-color);\n","      border-bottom-color: var(--fill-color);\n","    }\n","    90% {\n","      border-color: transparent;\n","      border-bottom-color: var(--fill-color);\n","    }\n","  }\n","</style>\n","\n","  <script>\n","    async function quickchart(key) {\n","      const quickchartButtonEl =\n","        document.querySelector('#' + key + ' button');\n","      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n","      quickchartButtonEl.classList.add('colab-df-spinner');\n","      try {\n","        const charts = await google.colab.kernel.invokeFunction(\n","            'suggestCharts', [key], {});\n","      } catch (error) {\n","        console.error('Error during call to suggestCharts:', error);\n","      }\n","      quickchartButtonEl.classList.remove('colab-df-spinner');\n","      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n","    }\n","    (() => {\n","      let quickchartButtonEl =\n","        document.querySelector('#df-054463a8-10f5-4de4-aa23-b85ce1864d6b button');\n","      quickchartButtonEl.style.display =\n","        google.colab.kernel.accessAllowed ? 'block' : 'none';\n","    })();\n","  </script>\n","</div>\n","\n","    </div>\n","  </div>\n"]},"metadata":{},"execution_count":118}],"source":["genetic_data = get_genetic_data(matrix_file)\n","genetic_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590261560Z","start_time":"2023-12-31T21:19:35.493320362Z"},"colab":{"base_uri":"https://localhost:8080/","height":256},"id":"gVXr4nRrZWQb","executionInfo":{"status":"ok","timestamp":1704461906840,"user_tz":-480,"elapsed":488,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"7b4b3ed7-def6-4508-a363-dbeff6bc883b"}},{"cell_type":"code","execution_count":119,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['1007_s_at',\n"," '1053_at',\n"," '117_at',\n"," '121_at',\n"," '1255_g_at',\n"," '1294_at',\n"," '1316_at',\n"," '1320_at',\n"," '1405_i_at',\n"," '1431_at',\n"," '1438_at',\n"," '1487_at',\n"," '1494_f_at',\n"," '1598_g_at',\n"," '160020_at',\n"," '1729_at',\n"," '1773_at',\n"," '177_at',\n"," '179_at',\n"," '1861_at']"]},"metadata":{},"execution_count":119}],"source":["gene_row_ids = genetic_data['ID'][:20].tolist()\n","gene_row_ids"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590443507Z","start_time":"2023-12-31T21:19:35.536713559Z"},"colab":{"base_uri":"https://localhost:8080/"},"id":"j3WTypxBZWQc","executionInfo":{"status":"ok","timestamp":1704461909584,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"972480e2-804d-49e9-e709-eed5f3368e9d"}},{"cell_type":"markdown","source":["Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:30:41.595335164Z","start_time":"2023-12-31T03:30:41.513232329Z"},"id":"FrSj5umZZWQc"}},{"cell_type":"markdown","source":["Reference prompt:"],"metadata":{"collapsed":false,"id":"DpeO_Jd3ZWQc"}},{"cell_type":"code","execution_count":120,"outputs":[{"output_type":"execute_result","data":{"text/plain":["\"\\nBelow are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\\nrequires_gene_mapping = (True or False)\\n\\nRow headers:\\n['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at', '1294_at', '1316_at', '1320_at', '1405_i_at', '1431_at', '1438_at', '1487_at', '1494_f_at', '1598_g_at', '160020_at', '1729_at', '1773_at', '177_at', '179_at', '1861_at']\\n\""],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":120}],"source":["f'''\n","Below are the row headers of a gene expression dataset in GEO. Based on your biomedical knowledge, are they human gene symbols, or are they some other identifiers that need to be mapped to gene symbols? Your answer should be concluded by starting a new line and strictly following this format:\n","requires_gene_mapping = (True or False)\n","\n","Row headers:\n","{gene_row_ids}\n","'''"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590590427Z","start_time":"2023-12-31T21:19:35.536780517Z"},"colab":{"base_uri":"https://localhost:8080/","height":107},"id":"kInJXlliZWQc","executionInfo":{"status":"ok","timestamp":1704461911916,"user_tz":-480,"elapsed":4,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"1fb3f4f4-8fad-438e-97d4-e9607981c081"}},{"cell_type":"markdown","source":["If not required, jump directly to the gene normalization step"],"metadata":{"collapsed":false,"id":"wEp4DKmcZWQc"}},{"cell_type":"code","execution_count":121,"outputs":[],"source":["requires_gene_mapping = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590648558Z","start_time":"2023-12-31T21:19:35.536821375Z"},"id":"kv3cD6jRZWQc","executionInfo":{"status":"ok","timestamp":1704461913878,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":122,"outputs":[],"source":["if requires_gene_mapping:\n","    gene_annotation = get_gene_annotation(soft_file)[0]\n","    gene_annotation_summary = preview_df(gene_annotation)\n","    gene_annotation_summary"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590695457Z","start_time":"2023-12-31T21:19:35.536910836Z"},"id":"TBTqW54jZWQc","executionInfo":{"status":"ok","timestamp":1704461920102,"user_tz":-480,"elapsed":4193,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["Observe the first few cells in the ID column of the gene annotation dataframe, to find the names of columns that store the gene probe IDs and gene symbols respectively.\n","Reference prompt:"],"metadata":{"collapsed":false,"ExecuteTime":{"end_time":"2023-12-31T03:58:04.978931927Z","start_time":"2023-12-31T03:58:04.966328339Z"},"id":"Lg3BW_bsZWQc"}},{"cell_type":"code","execution_count":123,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n","    \n"]}],"source":["if requires_gene_mapping:\n","    print(f'''\n","    As a biomedical research team, we extracted the gene annotation data from a series in the GEO database, and saved it to a Python dictionary. Please read the dictionary, and decide which key stores the ID of the probe, and which key stores the gene symbols. Please strict follow this format in your answer:\n","    probe_name_key = key_name1\n","    gene_name_key = key_name2\n","\n","    Gene annotation dictionary:\n","    {gene_annotation_summary}\n","    ''')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590742236Z","start_time":"2023-12-31T21:19:35.537050893Z"},"id":"kY03dnyLZWQc","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704461925223,"user_tz":-480,"elapsed":496,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"b294f806-e6bc-4dcb-cc06-a685c7ef3ec2"}},{"cell_type":"code","execution_count":124,"outputs":[],"source":["if requires_gene_mapping:\n","    probe_id_key = 'ID'\n","    gene_symb_key = 'Gene Symbol'\n","    gene_mapping = get_gene_mapping(gene_annotation, probe_id_key, gene_symb_key)\n","    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590786881Z","start_time":"2023-12-31T21:19:35.537132859Z"},"id":"fD2yNE8vZWQd","executionInfo":{"status":"ok","timestamp":1704461941204,"user_tz":-480,"elapsed":631,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":125,"outputs":[{"output_type":"stream","name":"stderr","text":["INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:9 input query terms found dup hits:\t[('ALDOAP2', 2), ('ALOX12P2', 2), ('ANKRD10-IT1', 2), ('ANKRD36BP2', 2), ('ANXA2P1', 2), ('ANXA2P2',\n","WARNING:biothings.client:143 input query terms found no hit:\t['AARS', 'AARSD1 /// PTGES3L /// PTGES3L-AARSD1', 'ABCB1 /// ABCB4', 'ABCB8 /// ASIC3', 'ABCC6 /// A\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:10 input query terms found dup hits:\t[('ATXN8OS', 2), ('AZGP1P1', 2), ('BRD7P3', 2), ('BTF3P11', 2), ('BTN2A3P', 2), ('C2orf27A', 2), ('C\n","WARNING:biothings.client:191 input query terms found no hit:\t['AURKAPS1 /// RAB3GAP2', 'AV8S2 /// TRAV13-2 /// TRAV13-2', 'AV8S2A1N1T /// hADV36S1 /// TCR-alpha \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:11 input query terms found dup hits:\t[('CELP', 2), ('CLUHP3', 2), ('CMAHP', 2), ('CNTNAP3P2', 2), ('COPG2IT1', 2), ('CROCCP2', 2), ('CRYB\n","WARNING:biothings.client:108 input query terms found no hit:\t['CDRT1', 'CDY1 /// CDY1B /// CDY2A /// CDY2B', 'CEBPA-AS1', 'CECR1', 'CECR5', 'CELA2A /// CELA2B', \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('DLEU2L', 2), ('DPY19L1P1', 2), ('DPY19L2P2', 3), ('DSTNP2', 2), ('EEF1A1P42', 2), ('ELK2AP', 3), \n","WARNING:biothings.client:125 input query terms found no hit:\t['DDX58', 'DEC1', 'DEFA1 /// DEFA1B /// DEFA3', 'DEFB4A /// DEFB4B', 'DES /// SUPT20H', 'DFNA5', 'DF\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:11 input query terms found dup hits:\t[('FCGR2C', 2), ('FER1L4', 2), ('FOLH1B', 2), ('FTH1P5', 2), ('GABARAPL3', 2), ('GOLGA2P5', 2), ('GT\n","WARNING:biothings.client:125 input query terms found no hit:\t['FCF1 /// MAPK1IP1L', 'FCGR1A /// FCGR1B /// FCGR1C', 'FCGR1B', 'FCGR3A /// FCGR3B', 'FDPSP5 /// FD\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('HLA-DRB6', 2), ('HLA-J', 9), ('HMGB3P1', 2), ('HNRNPA3P1', 2), ('ID2B', 2), ('IGHD', 3), ('IGHG1'\n","WARNING:biothings.client:183 input query terms found no hit:\t['HEATR2', 'HERC2 /// HERC2P10 /// LOC101929047 /// LOC101929832', 'HERC2P2 /// HERC2P9', 'HERC2P3 /\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:11 input query terms found dup hits:\t[('KIR3DX1', 5), ('KLF3-AS1', 2), ('KRT8P12', 2), ('LINC00115', 2), ('LINC00342', 2), ('LINC00652', \n","WARNING:biothings.client:291 input query terms found no hit:\t['KIAA0100', 'KIAA0101', 'KIAA0125', 'KIAA0141', 'KIAA0195 /// MIR6785', 'KIAA0196', 'KIAA0226', 'KI\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:6 input query terms found dup hits:\t[('MCF2L-AS1', 2), ('MEIS3P1', 2), ('MKRN7P', 2), ('MST1L', 2), ('NAT8B', 3), ('NOP14-AS1', 2)]\n","WARNING:biothings.client:119 input query terms found no hit:\t['MBTPS2 /// YY2', 'MCAM /// MIR6756', 'MCFD2P1 /// MCFD2P1', 'MED24 /// MIR6884', 'MEF2B /// MEF2BN\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:15 input query terms found dup hits:\t[('NUS1P3', 2), ('OR7E12P', 2), ('OR7E14P', 2), ('OR7E156P', 2), ('OR7E19P', 2), ('OR7E2P', 2), ('OR\n","WARNING:biothings.client:63 input query terms found no hit:\t['NOTCH2NL', 'NOV', 'NPIPA1 /// NPIPA5 /// NPIPB15 /// NPIPB3 /// NPIPB6 /// PDXDC2P', 'NPIPA5 /// N\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:14 input query terms found dup hits:\t[('POM121L9P', 2), ('POTEKP', 2), ('PPBPP2', 2), ('PPIEL', 2), ('PRORY', 2), ('PRR34', 2), ('PRSS3P2\n","WARNING:biothings.client:70 input query terms found no hit:\t['PNMAL1', 'POLDIP3 /// RRP7B', 'POLR2J /// POLR2J2 /// POLR2J3 /// POLR2J4 /// UPK3BL', 'POLR2J ///\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:8 input query terms found dup hits:\t[('RNF185-AS1', 2), ('RPL23AP32', 2), ('RPL23AP53', 2), ('RRN3P1', 2), ('SAA3P', 2), ('SCAND2P', 3),\n","WARNING:biothings.client:176 input query terms found no hit:\t['RNF11B', 'RNF219', 'RNF5 /// RNF5P1', 'RNFT1 /// TBC1D3P1-DHX40P1', 'RNMTL1', 'RNU86 /// RPL3 /// \n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:10 input query terms found dup hits:\t[('SMG7-AS1', 2), ('SNORA21', 2), ('SNX29P2', 2), ('SP3P', 2), ('SPRR2C', 2), ('ST20', 2), ('STAG3L3\n","WARNING:biothings.client:69 input query terms found no hit:\t['SLC6A10P /// SLC6A10PB /// SLC6A8', 'SLC9A3R1', 'SLC9A3R2', 'SLMO1', 'SLMO2', 'SMA4', 'SMARCE1P2 /\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-1000...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:12 input query terms found dup hits:\t[('TP53TG1', 2), ('TPT1P8', 2), ('TRAC', 2), ('TRAV12-2', 2), ('TRAV8-3', 2), ('TRBC1', 3), ('TRDV3'\n","WARNING:biothings.client:89 input query terms found no hit:\t['TMED7 /// TMED7-TICAM2', 'TMEM110', 'TMEM159', 'TMEM180', 'TMEM183A /// TMEM183B', 'TMEM189 /// TM\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n","INFO:biothings.client:querying 1-484...\n","INFO:biothings.client:done.\n","INFO:biothings.client:Finished.\n","WARNING:biothings.client:5 input query terms found dup hits:\t[('WT1-AS', 2), ('ZNF137P', 2), ('ZNF204P', 2), ('ZNF213-AS1', 2), ('ZNF767P', 2)]\n","WARNING:biothings.client:28 input query terms found no hit:\t['WRB', 'XAGE1B /// XAGE1E', 'XCL1 /// XCL2', 'XKRY /// XKRY2', 'Y16709', 'YARS', 'ZAK', 'ZBED8', 'Z\n","INFO:biothings.client:Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n"]}],"source":["if NORMALIZE_GENE:\n","    genetic_data = normalize_gene_symbols_in_index(genetic_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590837318Z","start_time":"2023-12-31T21:19:35.537209586Z"},"id":"IYTbjGQJZWQd","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704461980011,"user_tz":-480,"elapsed":35961,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}},"outputId":"7cd49ab0-0f06-4ce0-fc7b-6b338d6ffcf6"}},{"cell_type":"code","execution_count":126,"outputs":[],"source":["merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n","# The preprocessing runs through, which means is_available should be True\n","is_available = True"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.590902191Z","start_time":"2023-12-31T21:19:35.537285901Z"},"id":"gGEX1XUUZWQd","executionInfo":{"status":"ok","timestamp":1704462005405,"user_tz":-480,"elapsed":479,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":127,"outputs":[{"output_type":"stream","name":"stdout","text":["The merged dataset contains 32 samples.\n"]}],"source":["print(f\"The merged dataset contains {len(merged_data)} samples.\")"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.591095170Z","start_time":"2023-12-31T21:19:35.538708445Z"},"id":"Iy9Qpg0MZWQd","outputId":"b8d07666-8fdb-4b0a-e9d4-5ba48e2a2ab7","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704462007515,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":128,"outputs":[{"output_type":"stream","name":"stdout","text":["For the feature 'Lower Grade Glioma', the least common label is '1' with 32 occurrences. This represents 100.00% of the dataset.\n","The distribution of the feature 'Lower Grade Glioma' in this dataset is severely biased.\n","\n","Quartiles for 'Age':\n","  25%: 49.75\n","  50% (Median): 65.5\n","  75%: 70.0\n","Min: 30\n","Max: 76\n","The distribution of the feature 'Age' in this dataset is fine.\n","\n","For the feature 'Gender', the least common label is '1' with 13 occurrences. This represents 40.62% of the dataset.\n","The distribution of the feature 'Gender' in this dataset is fine.\n","\n"]},{"output_type":"execute_result","data":{"text/plain":["True"]},"metadata":{},"execution_count":128}],"source":["is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n","is_trait_biased"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610849182Z","start_time":"2023-12-31T21:19:35.541210258Z"},"id":"LXD45pSgZWQd","outputId":"707fbec1-1060-4f12-d27e-dafa6717731e","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1704462009710,"user_tz":-480,"elapsed":2,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":129,"outputs":[],"source":["save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610923234Z","start_time":"2023-12-31T21:19:35.584848473Z"},"id":"OZ0SAESiZWQd","executionInfo":{"status":"ok","timestamp":1704462016358,"user_tz":-480,"elapsed":509,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"code","execution_count":130,"outputs":[],"source":["merged_data.head()\n","if not is_trait_biased:\n","    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort+'.csv'), index=False)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:35.610970734Z","start_time":"2023-12-31T21:19:35.585042694Z"},"id":"02hAGq1wZWQd","executionInfo":{"status":"ok","timestamp":1704462019117,"user_tz":-480,"elapsed":472,"user":{"displayName":"郭淑怡","userId":"04547554131744612440"}}}},{"cell_type":"markdown","source":["### 3. Do regression & Cross Validation"],"metadata":{"id":"-MTPhRGxJV7I","ExecuteTime":{"end_time":"2023-12-31T03:58:22.271683755Z","start_time":"2023-12-31T03:58:22.246557674Z"}}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"'Xena'"},"execution_count":44,"metadata":{},"output_type":"execute_result"}],"source":["best_cohort, ranked_df = filter_and_rank_cohorts(JSON_PATH, 'has_age')\n","best_cohort"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.459442087Z","start_time":"2023-12-31T21:19:44.416219648Z"},"id":"2P5jb3pNZWQd","outputId":"b8aac6c6-600d-46d7-d8dd-12ff49ad4a04"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"   cohort_id  is_usable  is_available is_biased has_age has_gender  \\\n0       Xena       True          True     False    True       True   \n3  GSE248830       True          True     False    True       True   \n\n   sample_size note  \n0       1215.0       \n3         38.0       ","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>cohort_id</th>\n      <th>is_usable</th>\n      <th>is_available</th>\n      <th>is_biased</th>\n      <th>has_age</th>\n      <th>has_gender</th>\n      <th>sample_size</th>\n      <th>note</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>Xena</td>\n      <td>True</td>\n      <td>True</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>1215.0</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>GSE248830</td>\n      <td>True</td>\n      <td>True</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>38.0</td>\n      <td></td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"execution_count":45,"metadata":{},"output_type":"execute_result"}],"source":["ranked_df.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:44.808945960Z","start_time":"2023-12-31T21:19:44.799467072Z"},"id":"ooo1W8o8ZWQd","outputId":"a6b59e69-83c9-4659-9389-949704b9c191"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["merged_data = pd.read_csv(os.path.join(OUTPUT_DIR, best_cohort+'.csv'))"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.222719978Z","start_time":"2023-12-31T21:19:45.162873006Z"},"id":"4_XKlkP_ZWQd"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"   Breast Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17     RNF10  \\\n0              1  55.0     0.0   0.607308 -3.194126 -0.531035 -0.145872   \n1              1  50.0     0.0  -0.641192 -4.928226  0.095465  0.098128   \n2              1  62.0     0.0   1.082808 -4.623726 -0.531035  0.484028   \n3              1  52.0     0.0   0.121608 -2.881526 -0.531035  0.179128   \n4              1  50.0     0.0   0.420208 -3.282726 -0.531035 -0.020972   \n\n      RNF11    RNF13   GTF2IP1  ...     TULP2     NPY5R     GNGT2    GNGT1  \\\n0  0.237422 -0.29921 -0.142694  ... -0.748878  0.565583 -0.767233 -1.28139   \n1 -0.541978 -0.32291 -0.044694  ...  0.312922  0.079683  0.475267  1.99661   \n2 -0.183678 -0.91901  0.261106  ... -0.748878 -0.656117 -0.216733  0.21081   \n3  0.039222 -0.45491  0.180306  ...  0.415422 -0.178417 -0.211233 -1.28139   \n4 -0.117978 -0.55781 -0.173794  ...  0.685222 -0.698717 -0.484233 -1.28139   \n\n      TULP3      PTRF     BCL6B     GSTK1      SELP      SELS  \n0 -0.271377 -0.492286  0.360373  1.067905  0.076267 -0.392212  \n1  0.161423  0.273714  0.644673 -0.428695  0.068667 -0.043812  \n2  0.101023  0.660514  1.295073  0.915105  0.168567  0.047788  \n3  0.108023  1.132814  0.496773  0.240105  3.099767  0.112888  \n4 -0.223577  1.222714  0.225573 -0.301995  0.200067 -0.080212  \n\n[5 rows x 20533 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Breast Cancer</th>\n      <th>Age</th>\n      <th>Gender</th>\n      <th>ARHGEF10L</th>\n      <th>HIF3A</th>\n      <th>RNF17</th>\n      <th>RNF10</th>\n      <th>RNF11</th>\n      <th>RNF13</th>\n      <th>GTF2IP1</th>\n      <th>...</th>\n      <th>TULP2</th>\n      <th>NPY5R</th>\n      <th>GNGT2</th>\n      <th>GNGT1</th>\n      <th>TULP3</th>\n      <th>PTRF</th>\n      <th>BCL6B</th>\n      <th>GSTK1</th>\n      <th>SELP</th>\n      <th>SELS</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>1</td>\n      <td>55.0</td>\n      <td>0.0</td>\n      <td>0.607308</td>\n      <td>-3.194126</td>\n      <td>-0.531035</td>\n      <td>-0.145872</td>\n      <td>0.237422</td>\n      <td>-0.29921</td>\n      <td>-0.142694</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>0.565583</td>\n      <td>-0.767233</td>\n      <td>-1.28139</td>\n      <td>-0.271377</td>\n      <td>-0.492286</td>\n      <td>0.360373</td>\n      <td>1.067905</td>\n      <td>0.076267</td>\n      <td>-0.392212</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>-0.641192</td>\n      <td>-4.928226</td>\n      <td>0.095465</td>\n      <td>0.098128</td>\n      <td>-0.541978</td>\n      <td>-0.32291</td>\n      <td>-0.044694</td>\n      <td>...</td>\n      <td>0.312922</td>\n      <td>0.079683</td>\n      <td>0.475267</td>\n      <td>1.99661</td>\n      <td>0.161423</td>\n      <td>0.273714</td>\n      <td>0.644673</td>\n      <td>-0.428695</td>\n      <td>0.068667</td>\n      <td>-0.043812</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>1</td>\n      <td>62.0</td>\n      <td>0.0</td>\n      <td>1.082808</td>\n      <td>-4.623726</td>\n      <td>-0.531035</td>\n      <td>0.484028</td>\n      <td>-0.183678</td>\n      <td>-0.91901</td>\n      <td>0.261106</td>\n      <td>...</td>\n      <td>-0.748878</td>\n      <td>-0.656117</td>\n      <td>-0.216733</td>\n      <td>0.21081</td>\n      <td>0.101023</td>\n      <td>0.660514</td>\n      <td>1.295073</td>\n      <td>0.915105</td>\n      <td>0.168567</td>\n      <td>0.047788</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>1</td>\n      <td>52.0</td>\n      <td>0.0</td>\n      <td>0.121608</td>\n      <td>-2.881526</td>\n      <td>-0.531035</td>\n      <td>0.179128</td>\n      <td>0.039222</td>\n      <td>-0.45491</td>\n      <td>0.180306</td>\n      <td>...</td>\n      <td>0.415422</td>\n      <td>-0.178417</td>\n      <td>-0.211233</td>\n      <td>-1.28139</td>\n      <td>0.108023</td>\n      <td>1.132814</td>\n      <td>0.496773</td>\n      <td>0.240105</td>\n      <td>3.099767</td>\n      <td>0.112888</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>1</td>\n      <td>50.0</td>\n      <td>0.0</td>\n      <td>0.420208</td>\n      <td>-3.282726</td>\n      <td>-0.531035</td>\n      <td>-0.020972</td>\n      <td>-0.117978</td>\n      <td>-0.55781</td>\n      <td>-0.173794</td>\n      <td>...</td>\n      <td>0.685222</td>\n      <td>-0.698717</td>\n      <td>-0.484233</td>\n      <td>-1.28139</td>\n      <td>-0.223577</td>\n      <td>1.222714</td>\n      <td>0.225573</td>\n      <td>-0.301995</td>\n      <td>0.200067</td>\n      <td>-0.080212</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 20533 columns</p>\n</div>"},"execution_count":47,"metadata":{},"output_type":"execute_result"}],"source":["merged_data.head()"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.234945878Z","start_time":"2023-12-31T21:19:52.231604618Z"},"id":"vTvOiZMsZWQe","outputId":"5c8cc9ed-391c-45f3-cf9a-a97bd59764f6"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["# If both age and gender features are available, select 'age' as the condition.\n","condition = 'Age'\n","# Remove the other condition to prevent interference.\n","merged_data = merged_data.drop(columns=['Gender'], errors='ignore').astype('float')"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.312540150Z","start_time":"2023-12-31T21:19:52.235772334Z"},"id":"M-4hO0PHZWQe"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["X = merged_data.drop(columns=[TRAIT, condition]).values\n","Y = merged_data[TRAIT].values\n","Z = merged_data[condition].values"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:19:52.323468266Z","start_time":"2023-12-31T21:19:52.290903989Z"},"id":"dhUrhPcqZWQe"}},{"cell_type":"markdown","source":["Select the appropriate regression model depending on whether the dataset shows batch effect."],"metadata":{"collapsed":false,"id":"u5Mx3XpAZWQf"}},{"cell_type":"code","execution_count":null,"outputs":[{"data":{"text/plain":"True"},"execution_count":50,"metadata":{},"output_type":"execute_result"}],"source":["has_batch_effect = detect_batch_effect(X)\n","has_batch_effect"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.930840782Z","start_time":"2023-12-31T21:19:52.319092572Z"},"id":"GdxnyJr4ZWQg","outputId":"69b5d731-d891-4d1b-dda9-bf4aa5156ad0"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":["if has_batch_effect:\n","    model_constructor = VariableSelection\n","    model_params = {}\n","else:\n","    model_constructor = Lasso\n","    model_params = {'alpha': 1.0, 'random_state': 42}"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:00.931029722Z","start_time":"2023-12-31T21:20:00.929710828Z"},"id":"uzy8ltp-ZWQg"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stderr","output_type":"stream","text":["/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n","/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n"]},{"name":"stdout","output_type":"stream","text":["The cross-validation accuracy is 91.28% ± 1.50%\n"]}],"source":["cv_mean, cv_std = cross_validation(X, Y, Z, model_constructor, model_params, target_type=trait_type)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:20.709459824Z","start_time":"2023-12-31T21:20:00.931657949Z"},"id":"z9tMro7zZWQg","outputId":"511e2df7-1da4-4654-e935-30c8c1ff3ef0"}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stderr","output_type":"stream","text":["/home/techt/anaconda3/envs/hl/lib/python3.8/site-packages/sparse_lmm/helpingMethods.py:49: RuntimeWarning: invalid value encountered in divide\n","  ts = beta / np.sqrt(var * sigma)\n"]}],"source":["# Train regression model on the whole dataset to identify significant genes\n","model = ResidualizationRegressor(model_constructor, model_params)\n","normalized_X, _ = normalize_data(X)\n","normalized_Z, _ = normalize_data(Z)\n","model.fit(normalized_X, Y, normalized_Z)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:24.980870159Z","start_time":"2023-12-31T21:20:20.516710807Z"},"id":"-GzlAnQWZWQg","outputId":"795203e6-3de7-4bed-ad39-060e287c69bd"}},{"cell_type":"markdown","source":["### 5. Discussion and report"],"metadata":{"id":"EjJrxbvb4nlj","ExecuteTime":{"end_time":"2023-10-14T04:50:59.072075075Z","start_time":"2023-10-14T04:50:38.739499998Z"}}},{"cell_type":"code","execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["Effect of the condition on the target variable:\n","Variable: Age\n","Coefficient: 0.0065\n","p-value: nan\n","\n","Found 76 significant genes associated with the trait 'Breast Cancer' conditional on the factor 'Age', with corrected p-value < 0.05:\n","    Variable  Coefficient  corrected_p_value\n","     COL10A1     0.136339       3.404090e-27\n","         BGN     0.095422       1.004817e-14\n","    PPAPDC1A     0.081296       9.785054e-13\n","       MMP11     0.080354       1.123019e-10\n","       WISP1     0.074888       9.613830e-10\n","     HSD17B6     0.051940       6.980811e-08\n","      CTHRC1     0.058310       2.210553e-07\n","       MFAP5    -0.054250       2.706783e-07\n","      CLEC5A     0.047846       8.877914e-07\n","   C20orf103     0.048957       8.877914e-07\n","          GC     0.027375       1.419502e-05\n","      OR51D1     0.022908       1.644207e-05\n","        COMP     0.041129       4.366228e-05\n","LOC100169752     0.019398       4.366228e-05\n","      LRRC15     0.052746       4.494130e-05\n","         FN1     0.063527       5.027562e-05\n","     C1QTNF6     0.049130       1.220626e-04\n","       INHBA     0.052621       1.384526e-04\n","       FNDC1     0.044080       1.724540e-04\n","        VCAN     0.062919       3.541404e-04\n","   LOC283867     0.039438       3.541404e-04\n","     SLC24A2     0.045359       3.929641e-04\n","        ECM2    -0.066303       4.895350e-04\n","       CILP2     0.040613       4.895350e-04\n","         DCN    -0.062282       5.405761e-04\n","        HEYL     0.039376       6.133304e-04\n","         FAP     0.062297       6.903458e-04\n","     OLFML2A    -0.037988       8.815285e-04\n","       MFAP2     0.047162       8.815285e-04\n","       CCRL1    -0.036802       8.815285e-04\n","       CD248    -0.046771       8.815285e-04\n","       P4HA3     0.044037       8.815285e-04\n","     C5orf13     0.037746       1.590045e-03\n","      COL5A2     0.064118       2.057973e-03\n","         F2R     0.041164       2.101197e-03\n","      OLFML1    -0.061570       3.573228e-03\n","       PRDM8    -0.040662       3.786028e-03\n","   POM121L12     0.028947       3.941936e-03\n","       LOXL1     0.044681       4.547038e-03\n","       SPARC     0.053330       5.391493e-03\n","      LEFTY2     0.027338       5.753501e-03\n","    SERPINE1     0.028865       5.753501e-03\n","       OR2L3    -0.016537       6.402547e-03\n","     CYP19A1     0.022029       6.651093e-03\n","         MAF    -0.038952       7.098913e-03\n","       GNG11    -0.068155       1.242228e-02\n","       CFHR5    -0.018470       1.366939e-02\n","      CLEC3B    -0.043039       1.366939e-02\n","        NRG3    -0.019375       1.366939e-02\n","      COL1A1     0.046342       1.427700e-02\n","      PCOLCE    -0.049663       1.836229e-02\n","     TMEM119     0.037238       1.922482e-02\n","       KCNC1    -0.024520       2.248216e-02\n","       WNT8A     0.014598       2.592754e-02\n","        RECK    -0.052887       2.781665e-02\n","        SNCB     0.018240       3.140935e-02\n","      KBTBD6    -0.029237       3.140935e-02\n","       SRPX2    -0.034794       3.140935e-02\n","     OLFML2B     0.036407       3.214194e-02\n","      PLXDC1    -0.032863       3.219694e-02\n","      ADAM12     0.040368       3.309987e-02\n","      GRIN2A    -0.021383       3.340752e-02\n","      COL5A1     0.050110       3.456174e-02\n","      USHBP1    -0.050144       3.618703e-02\n","       CYLC2    -0.016414       3.622489e-02\n","     TMEM90B     0.034482       3.622489e-02\n","      RESP18    -0.014081       3.622489e-02\n","        RGS4     0.022820       3.855158e-02\n","        CNTF    -0.020244       3.880001e-02\n","    ADAMTSL2     0.030360       4.155675e-02\n","        CST2     0.019834       4.218502e-02\n","       GRID1    -0.028724       4.218502e-02\n","       MATN3     0.028706       4.316484e-02\n","       GGT8P    -0.017000       4.494972e-02\n","      GPR101     0.014064       4.913174e-02\n","      KIF26B     0.031349       4.960098e-02\n"]},{"name":"stderr","output_type":"stream","text":["/home/techt/AI4Science_gold_standard/stage1/Haoyang/utils.py:492: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame.\n","Try using .loc[row_indexer,col_indexer] = value instead\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  gene_regression_df.loc[:, 'corrected_p_value'] = corrected_p_values\n"]}],"source":["feature_cols = merged_data.columns.tolist()\n","feature_cols.remove(TRAIT)\n","\n","if has_batch_effect:\n","    report_result_from_lmm(model, feature_cols, TRAIT, condition, threshold=0.05, save_output=True,\n","                           output_dir=OUTPUT_DIR)\n","else:\n","    report_result_from_lasso(model, feature_cols, TRAIT, condition, save_output=True, output_dir=OUTPUT_DIR)"],"metadata":{"ExecuteTime":{"end_time":"2023-12-31T21:20:25.000830410Z","start_time":"2023-12-31T21:20:24.982834541Z"},"id":"FHjsmuPhZWQg","outputId":"72e95ed5-03fc-49b2-ff5b-f014c8585088"}},{"cell_type":"code","execution_count":null,"outputs":[],"source":[],"metadata":{"id":"Ia26hwCEZWQg"}}]}