{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = 'Cervical Cancer'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:21.140790869Z",
     "start_time": "2024-02-27T00:22:21.133208795Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Cervical_Cancer_(CESC)',\n",
       " 'TCGA_Kidney_Chromophobe_(KICH)',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)',\n",
       " 'TCGA_Lower_Grade_Glioma_(LGG)',\n",
       " 'TCGA_Melanoma_(SKCM)',\n",
       " 'TCGA_Mesothelioma_(MESO)',\n",
       " 'TCGA_Testicular_Cancer_(TGCT)',\n",
       " 'TCGA_Uterine_Carcinosarcoma_(UCS)']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:21.985144212Z",
     "start_time": "2024-02-27T00:22:21.981084968Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\TCGA\\\\TCGA_Cervical_Cancer_(CESC)\\\\TCGA.CESC.sampleMap_CESC_clinicalMatrix',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\TCGA\\\\TCGA_Cervical_Cancer_(CESC)\\\\TCGA.CESC.sampleMap_HiSeqV2_PANCAN.gz')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_subdir = \"TCGA_Cervical_Cancer_(CESC)\"\n",
    "cohort = 'Xena'\n",
    "trait_type = 'binary'\n",
    "is_available = True\n",
    "\n",
    "cohort_dir = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "clinical_data_file, genetic_data_file = xena_get_relevant_filepaths(cohort_dir)\n",
    "clinical_data_file, genetic_data_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.794819905Z",
     "start_time": "2024-02-27T00:22:22.357077391Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.798785383Z",
     "start_time": "2024-02-27T00:22:25.796474039Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'additional_pharmaceutical_therapy',\n",
       " 'additional_radiation_therapy',\n",
       " 'additional_treatment_completion_success_outcome',\n",
       " 'adjuvant_rad_therapy_prior_admin',\n",
       " 'age_at_initial_pathologic_diagnosis']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.802838875Z",
     "start_time": "2024-02-27T00:22:25.798659327Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'additional_treatment_completion_success_outcome', 'adjuvant_rad_therapy_prior_admin', 'age_at_initial_pathologic_diagnosis', 'age_began_smoking_in_years', 'agent_total_dose_count', 'assessment_timepoint_category', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'birth_control_pill_history_usage_category', 'brachytherapy_administered_status', 'brachytherapy_first_reference_point_administered_total_dose', 'brachytherapy_method_other_specify_text', 'brachytherapy_method_type', 'cervical_carcinoma_corpus_uteri_involvement_indicator', 'cervical_carcinoma_pelvic_extension_text', 'cervical_neoplasm_pathologic_margin_involved_text', 'cervical_neoplasm_pathologic_margin_involved_type', 'chemotherapy_negation_radiation_therapy_concurrent_adminstrd_txt', 'chemotherapy_negation_radiation_therapy_concurrnt_nt_dmnstrd_rsn', 'chemotherapy_regimen_type', 'clinical_stage', 'concurrent_chemotherapy_dose', 'days_to_birth', 'days_to_brachytherapy_begin_occurrence', 'days_to_brachytherapy_end_occurrence', 'days_to_chemotherapy_end', 'days_to_chemotherapy_start', 'days_to_collection', 'days_to_death', 'days_to_diagnostic_computed_tomography_performed', 'days_to_diagnostic_mri_performed', 'days_to_fdg_or_ct_pet_performed', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'days_to_performance_status_assessment', 'days_to_radiation_therapy_end', 'days_to_radiation_therapy_start', 'death_cause_text', 'diagnostic_ct_result_outcome', 'diagnostic_mri_result_outcome', 'dose_frequency_text', 'eastern_cancer_oncology_group', 'ectopic_pregnancy_count', 'external_beam_radiation_therapy_administered_status', 'external_beam_radiation_therapy_administrd_prrtc_rgn_lymph_nd_ds', 'fdg_or_ct_pet_performed_outcome', 'female_breast_feeding_or_pregnancy_status_indicator', 'followup_case_report_form_submission_reason', 'form_completion_date', 'gender', 'height', 'histological_type', 'history_of_neoadjuvant_treatment', 'human_papillomavirus_laboratory_procedure_performed_name', 'human_papillomavirus_laboratory_procedure_performed_text', 'human_papillomavirus_other_type_text', 'human_papillomavirus_type', 'hysterectomy_performed_text', 'hysterectomy_performed_type', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'init_pathology_dx_method_other', 'initial_pathologic_diagnosis_method', 'initial_weight', 'is_ffpe', 'keratinizing_squamous_cell_carcinoma_present_indicator', 'lost_follow_up', 'lymph_node_examined_count', 'lymph_node_location_positive_pathology_name', 'lymph_node_location_positive_pathology_text', 'lymphovascular_invasion_indicator', 'menopause_status', 'neoplasm_histologic_grade', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_post_initial_therapy_diagnosis_method_text', 'new_neoplasm_event_post_initial_therapy_diagnosis_method_type', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'number_of_lymphnodes_positive_by_he', 'number_of_lymphnodes_positive_by_ihc', 'number_of_successful_pregnancies_which_resultd_n_t_lst_1_lv_brth', 'number_pack_years_smoked', 'oct_embedded', 'oligonucleotide_primer_pair_laboratory_procedure_performed_name', 'other_chemotherapy_agent_administration_specify', 'other_dx', 'pathologic_M', 'pathologic_N', 'pathologic_T', 'pathology_report_file_name', 'patient_death_reason', 'patient_history_immune_system_and_related_disorders_name', 'patient_history_immune_system_and_related_disorders_text', 'patient_id', 'patient_pregnancy_spontaneous_abortion_count', 'patient_pregnancy_therapeutic_abortion_count', 'performance_status_scale_timing', 'person_neoplasm_cancer_status', 'postoperative_rx_tx', 'pregnancy_stillbirth_count', 'primary_lymph_node_presentation_assessment', 'primary_therapy_outcome_success', 'radiation_therapy', 'radiation_therapy_not_administered_reason', 'radiation_therapy_not_administered_specify', 'radiation_type_notes', 'residual_disease_post_new_tumor_event_margin_status', 'rt_administered_type', 'rt_pelvis_administered_total_dose', 'sample_type', 'sample_type_id', 'standardized_uptake_value_cervix_uteri_assessment_measurement', 'stopped_smoking_year', 'system_version', 'targeted_molecular_therapy', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tobacco_smoking_history', 'total_number_of_pregnancies', 'tumor_response_cdus_type', 'tumor_tissue_site', 'vial_number', 'vital_status', 'weight', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_CESC_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_CESC_miRNA_HiSeq', '_GENOMIC_ID_TCGA_CESC_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_data/public/TCGA/CESC/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_CESC_PDMRNAseq', '_GENOMIC_ID_TCGA_CESC_RPPA', '_GENOMIC_ID_TCGA_CESC_hMethyl450', '_GENOMIC_ID_TCGA_CESC_mutation_bcgsc_gene', '_GENOMIC_ID_TCGA_CESC_exp_HiSeqV2_percentile', '_GENOMIC_ID_TCGA_CESC_mutation', '_GENOMIC_ID_TCGA_CESC_mutation_broad_gene', '_GENOMIC_ID_TCGA_CESC_mutation_ucsc_maf_gene', '_GENOMIC_ID_TCGA_CESC_mutation_curated_wustl_gene', '_GENOMIC_ID_TCGA_CESC_exp_HiSeqV2', '_GENOMIC_ID_TCGA_CESC_gistic2', '_GENOMIC_ID_TCGA_CESC_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_CESC_gistic2thd']\\n\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.864092575Z",
     "start_time": "2024-02-27T00:22:25.802001989Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = ['age_at_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = ['gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "\n",
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:26.108535524Z",
     "start_time": "2024-02-27T00:22:26.101435803Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_began_smoking_in_years'\n",
    "gender_col = 'gender'\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.602458026Z",
     "start_time": "2024-02-27T00:22:27.839636769Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-EA-A3HT-01</th>\n",
       "      <th>TCGA-FU-A3HZ-01</th>\n",
       "      <th>TCGA-FU-A3TQ-01</th>\n",
       "      <th>TCGA-DS-A0VK-01</th>\n",
       "      <th>TCGA-UC-A7PG-01</th>\n",
       "      <th>TCGA-C5-A7UE-01</th>\n",
       "      <th>TCGA-C5-A1MJ-01</th>\n",
       "      <th>TCGA-UC-A7PI-01</th>\n",
       "      <th>TCGA-C5-A1MF-01</th>\n",
       "      <th>TCGA-EA-A5ZF-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-JW-A5VI-01</th>\n",
       "      <th>TCGA-VS-A9UD-01</th>\n",
       "      <th>TCGA-C5-A1BQ-01</th>\n",
       "      <th>TCGA-MU-A8JM-01</th>\n",
       "      <th>TCGA-DS-A0VM-01</th>\n",
       "      <th>TCGA-MY-A5BF-01</th>\n",
       "      <th>TCGA-EA-A556-01</th>\n",
       "      <th>TCGA-IR-A3LA-01</th>\n",
       "      <th>TCGA-VS-AA62-01</th>\n",
       "      <th>TCGA-ZJ-AAXB-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <td>0.664808</td>\n",
       "      <td>-0.870192</td>\n",
       "      <td>1.278708</td>\n",
       "      <td>-0.210892</td>\n",
       "      <td>0.920108</td>\n",
       "      <td>-0.716092</td>\n",
       "      <td>0.054008</td>\n",
       "      <td>1.438008</td>\n",
       "      <td>-0.102892</td>\n",
       "      <td>-0.678292</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.108292</td>\n",
       "      <td>-0.458392</td>\n",
       "      <td>-0.413892</td>\n",
       "      <td>-1.007792</td>\n",
       "      <td>-0.549292</td>\n",
       "      <td>-0.255692</td>\n",
       "      <td>-1.172392</td>\n",
       "      <td>-0.146792</td>\n",
       "      <td>-1.306892</td>\n",
       "      <td>-0.492192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIF3A</th>\n",
       "      <td>-3.218726</td>\n",
       "      <td>-2.502826</td>\n",
       "      <td>-2.714626</td>\n",
       "      <td>-0.763726</td>\n",
       "      <td>-0.467326</td>\n",
       "      <td>-4.477926</td>\n",
       "      <td>-0.463926</td>\n",
       "      <td>-1.472326</td>\n",
       "      <td>1.105974</td>\n",
       "      <td>2.024374</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.011226</td>\n",
       "      <td>-2.142026</td>\n",
       "      <td>-2.121626</td>\n",
       "      <td>-4.116626</td>\n",
       "      <td>-3.194026</td>\n",
       "      <td>-4.054026</td>\n",
       "      <td>4.398574</td>\n",
       "      <td>-2.194926</td>\n",
       "      <td>-3.093226</td>\n",
       "      <td>0.063174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>0.577865</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.847665</td>\n",
       "      <td>2.327065</td>\n",
       "      <td>0.032765</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>1.457465</td>\n",
       "      <td>0.334265</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.930665</td>\n",
       "      <td>-0.123235</td>\n",
       "      <td>1.745865</td>\n",
       "      <td>1.218965</td>\n",
       "      <td>0.406265</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.054765</td>\n",
       "      <td>-0.531035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF10</th>\n",
       "      <td>-0.187472</td>\n",
       "      <td>-0.057972</td>\n",
       "      <td>-0.071672</td>\n",
       "      <td>-0.142172</td>\n",
       "      <td>0.148928</td>\n",
       "      <td>-0.214172</td>\n",
       "      <td>-0.073972</td>\n",
       "      <td>-0.196072</td>\n",
       "      <td>-0.199772</td>\n",
       "      <td>-0.428272</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.471072</td>\n",
       "      <td>-0.269972</td>\n",
       "      <td>0.048828</td>\n",
       "      <td>0.292028</td>\n",
       "      <td>-0.252872</td>\n",
       "      <td>-0.067372</td>\n",
       "      <td>0.193628</td>\n",
       "      <td>-0.223072</td>\n",
       "      <td>-0.040772</td>\n",
       "      <td>0.230928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF11</th>\n",
       "      <td>0.552722</td>\n",
       "      <td>-0.123378</td>\n",
       "      <td>1.481922</td>\n",
       "      <td>0.176122</td>\n",
       "      <td>0.509522</td>\n",
       "      <td>0.367822</td>\n",
       "      <td>-0.677078</td>\n",
       "      <td>-0.947978</td>\n",
       "      <td>-1.280178</td>\n",
       "      <td>-0.095978</td>\n",
       "      <td>...</td>\n",
       "      <td>0.503822</td>\n",
       "      <td>-0.587878</td>\n",
       "      <td>0.283922</td>\n",
       "      <td>0.592722</td>\n",
       "      <td>-1.018078</td>\n",
       "      <td>0.971822</td>\n",
       "      <td>0.041822</td>\n",
       "      <td>0.012922</td>\n",
       "      <td>-0.434478</td>\n",
       "      <td>-0.072878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNGT1</th>\n",
       "      <td>0.170210</td>\n",
       "      <td>1.443210</td>\n",
       "      <td>2.729510</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>0.002310</td>\n",
       "      <td>-0.644890</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>4.078610</td>\n",
       "      <td>-0.416090</td>\n",
       "      <td>...</td>\n",
       "      <td>1.661210</td>\n",
       "      <td>3.344110</td>\n",
       "      <td>3.013010</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>2.478010</td>\n",
       "      <td>-0.003090</td>\n",
       "      <td>1.421210</td>\n",
       "      <td>0.305710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TULP3</th>\n",
       "      <td>0.007523</td>\n",
       "      <td>-0.488077</td>\n",
       "      <td>0.415323</td>\n",
       "      <td>-0.644377</td>\n",
       "      <td>0.014623</td>\n",
       "      <td>-0.018677</td>\n",
       "      <td>-0.029077</td>\n",
       "      <td>-0.047377</td>\n",
       "      <td>-0.332877</td>\n",
       "      <td>0.111723</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020823</td>\n",
       "      <td>-0.019477</td>\n",
       "      <td>0.668923</td>\n",
       "      <td>0.274423</td>\n",
       "      <td>-0.220377</td>\n",
       "      <td>0.215323</td>\n",
       "      <td>0.248423</td>\n",
       "      <td>0.704523</td>\n",
       "      <td>-0.101977</td>\n",
       "      <td>-0.849077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL6B</th>\n",
       "      <td>-0.686627</td>\n",
       "      <td>-2.017227</td>\n",
       "      <td>-1.286527</td>\n",
       "      <td>0.938473</td>\n",
       "      <td>-1.135727</td>\n",
       "      <td>-2.895427</td>\n",
       "      <td>0.855573</td>\n",
       "      <td>-2.903127</td>\n",
       "      <td>-1.584927</td>\n",
       "      <td>-0.258327</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.775627</td>\n",
       "      <td>-1.650527</td>\n",
       "      <td>-1.165327</td>\n",
       "      <td>-0.596427</td>\n",
       "      <td>0.284873</td>\n",
       "      <td>0.572473</td>\n",
       "      <td>-0.361127</td>\n",
       "      <td>-2.012927</td>\n",
       "      <td>-0.930727</td>\n",
       "      <td>-0.373927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSTK1</th>\n",
       "      <td>-1.185895</td>\n",
       "      <td>-0.997295</td>\n",
       "      <td>-0.252695</td>\n",
       "      <td>0.038605</td>\n",
       "      <td>0.028005</td>\n",
       "      <td>-1.335795</td>\n",
       "      <td>0.054005</td>\n",
       "      <td>0.716605</td>\n",
       "      <td>0.648805</td>\n",
       "      <td>-0.636095</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.925495</td>\n",
       "      <td>0.676705</td>\n",
       "      <td>0.210605</td>\n",
       "      <td>-0.788095</td>\n",
       "      <td>0.490205</td>\n",
       "      <td>-1.347095</td>\n",
       "      <td>-0.309695</td>\n",
       "      <td>1.093805</td>\n",
       "      <td>0.746105</td>\n",
       "      <td>-0.192695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELP</th>\n",
       "      <td>-2.167333</td>\n",
       "      <td>-2.207133</td>\n",
       "      <td>0.967167</td>\n",
       "      <td>1.473667</td>\n",
       "      <td>-0.943233</td>\n",
       "      <td>-0.990433</td>\n",
       "      <td>1.256167</td>\n",
       "      <td>-0.677333</td>\n",
       "      <td>-0.040733</td>\n",
       "      <td>0.029467</td>\n",
       "      <td>...</td>\n",
       "      <td>0.926867</td>\n",
       "      <td>-1.312633</td>\n",
       "      <td>-4.128533</td>\n",
       "      <td>-1.249233</td>\n",
       "      <td>-1.127033</td>\n",
       "      <td>-3.046633</td>\n",
       "      <td>-2.762433</td>\n",
       "      <td>-1.405033</td>\n",
       "      <td>-0.435633</td>\n",
       "      <td>-2.715433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17339 rows × 308 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TCGA-EA-A3HT-01  TCGA-FU-A3HZ-01  TCGA-FU-A3TQ-01  TCGA-DS-A0VK-01  \\\n",
       "ARHGEF10L         0.664808        -0.870192         1.278708        -0.210892   \n",
       "HIF3A            -3.218726        -2.502826        -2.714626        -0.763726   \n",
       "RNF17             0.577865        -0.531035         1.847665         2.327065   \n",
       "RNF10            -0.187472        -0.057972        -0.071672        -0.142172   \n",
       "RNF11             0.552722        -0.123378         1.481922         0.176122   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             0.170210         1.443210         2.729510        -1.281390   \n",
       "TULP3             0.007523        -0.488077         0.415323        -0.644377   \n",
       "BCL6B            -0.686627        -2.017227        -1.286527         0.938473   \n",
       "GSTK1            -1.185895        -0.997295        -0.252695         0.038605   \n",
       "SELP             -2.167333        -2.207133         0.967167         1.473667   \n",
       "\n",
       "           TCGA-UC-A7PG-01  TCGA-C5-A7UE-01  TCGA-C5-A1MJ-01  TCGA-UC-A7PI-01  \\\n",
       "ARHGEF10L         0.920108        -0.716092         0.054008         1.438008   \n",
       "HIF3A            -0.467326        -4.477926        -0.463926        -1.472326   \n",
       "RNF17             0.032765        -0.531035        -0.531035        -0.531035   \n",
       "RNF10             0.148928        -0.214172        -0.073972        -0.196072   \n",
       "RNF11             0.509522         0.367822        -0.677078        -0.947978   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             0.002310        -0.644890        -1.281390        -1.281390   \n",
       "TULP3             0.014623        -0.018677        -0.029077        -0.047377   \n",
       "BCL6B            -1.135727        -2.895427         0.855573        -2.903127   \n",
       "GSTK1             0.028005        -1.335795         0.054005         0.716605   \n",
       "SELP             -0.943233        -0.990433         1.256167        -0.677333   \n",
       "\n",
       "           TCGA-C5-A1MF-01  TCGA-EA-A5ZF-01  ...  TCGA-JW-A5VI-01  \\\n",
       "ARHGEF10L        -0.102892        -0.678292  ...        -0.108292   \n",
       "HIF3A             1.105974         2.024374  ...        -3.011226   \n",
       "RNF17             1.457465         0.334265  ...        -0.531035   \n",
       "RNF10            -0.199772        -0.428272  ...        -0.471072   \n",
       "RNF11            -1.280178        -0.095978  ...         0.503822   \n",
       "...                    ...              ...  ...              ...   \n",
       "GNGT1             4.078610        -0.416090  ...         1.661210   \n",
       "TULP3            -0.332877         0.111723  ...         0.020823   \n",
       "BCL6B            -1.584927        -0.258327  ...        -0.775627   \n",
       "GSTK1             0.648805        -0.636095  ...        -0.925495   \n",
       "SELP             -0.040733         0.029467  ...         0.926867   \n",
       "\n",
       "           TCGA-VS-A9UD-01  TCGA-C5-A1BQ-01  TCGA-MU-A8JM-01  TCGA-DS-A0VM-01  \\\n",
       "ARHGEF10L        -0.458392        -0.413892        -1.007792        -0.549292   \n",
       "HIF3A            -2.142026        -2.121626        -4.116626        -3.194026   \n",
       "RNF17             0.930665        -0.123235         1.745865         1.218965   \n",
       "RNF10            -0.269972         0.048828         0.292028        -0.252872   \n",
       "RNF11            -0.587878         0.283922         0.592722        -1.018078   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             3.344110         3.013010        -1.281390        -1.281390   \n",
       "TULP3            -0.019477         0.668923         0.274423        -0.220377   \n",
       "BCL6B            -1.650527        -1.165327        -0.596427         0.284873   \n",
       "GSTK1             0.676705         0.210605        -0.788095         0.490205   \n",
       "SELP             -1.312633        -4.128533        -1.249233        -1.127033   \n",
       "\n",
       "           TCGA-MY-A5BF-01  TCGA-EA-A556-01  TCGA-IR-A3LA-01  TCGA-VS-AA62-01  \\\n",
       "ARHGEF10L        -0.255692        -1.172392        -0.146792        -1.306892   \n",
       "HIF3A            -4.054026         4.398574        -2.194926        -3.093226   \n",
       "RNF17             0.406265        -0.531035        -0.531035         0.054765   \n",
       "RNF10            -0.067372         0.193628        -0.223072        -0.040772   \n",
       "RNF11             0.971822         0.041822         0.012922        -0.434478   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -1.281390         2.478010        -0.003090         1.421210   \n",
       "TULP3             0.215323         0.248423         0.704523        -0.101977   \n",
       "BCL6B             0.572473        -0.361127        -2.012927        -0.930727   \n",
       "GSTK1            -1.347095        -0.309695         1.093805         0.746105   \n",
       "SELP             -3.046633        -2.762433        -1.405033        -0.435633   \n",
       "\n",
       "           TCGA-ZJ-AAXB-01  \n",
       "ARHGEF10L        -0.492192  \n",
       "HIF3A             0.063174  \n",
       "RNF17            -0.531035  \n",
       "RNF10             0.230928  \n",
       "RNF11            -0.072878  \n",
       "...                    ...  \n",
       "GNGT1             0.305710  \n",
       "TULP3            -0.849077  \n",
       "BCL6B            -0.373927  \n",
       "GSTK1            -0.192695  \n",
       "SELP             -2.715433  \n",
       "\n",
       "[17339 rows x 308 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.720144304Z",
     "start_time": "2024-02-27T00:23:16.643071320Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cervical Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-BI-A0VR-01</th>\n",
       "      <td>1</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.352292</td>\n",
       "      <td>-2.344926</td>\n",
       "      <td>0.592565</td>\n",
       "      <td>-0.256772</td>\n",
       "      <td>0.176622</td>\n",
       "      <td>-0.02901</td>\n",
       "      <td>-0.346094</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.270678</td>\n",
       "      <td>-1.587117</td>\n",
       "      <td>0.725467</td>\n",
       "      <td>0.08141</td>\n",
       "      <td>0.103323</td>\n",
       "      <td>-0.890027</td>\n",
       "      <td>-0.151595</td>\n",
       "      <td>0.767467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-C5-A1BI-01</th>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.403208</td>\n",
       "      <td>-2.758326</td>\n",
       "      <td>0.462565</td>\n",
       "      <td>0.011628</td>\n",
       "      <td>0.168622</td>\n",
       "      <td>-0.05601</td>\n",
       "      <td>-0.784594</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.016978</td>\n",
       "      <td>-1.587117</td>\n",
       "      <td>0.482567</td>\n",
       "      <td>0.29501</td>\n",
       "      <td>0.262323</td>\n",
       "      <td>-0.657127</td>\n",
       "      <td>-0.189895</td>\n",
       "      <td>0.948267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-C5-A1BJ-01</th>\n",
       "      <td>1</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.314392</td>\n",
       "      <td>-4.098426</td>\n",
       "      <td>2.206165</td>\n",
       "      <td>-0.411172</td>\n",
       "      <td>0.083522</td>\n",
       "      <td>-0.28401</td>\n",
       "      <td>-0.401694</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.562186</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>-0.682517</td>\n",
       "      <td>0.168067</td>\n",
       "      <td>2.51161</td>\n",
       "      <td>0.363123</td>\n",
       "      <td>0.377473</td>\n",
       "      <td>-0.686295</td>\n",
       "      <td>1.108467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-C5-A1BL-01</th>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.455892</td>\n",
       "      <td>-1.485426</td>\n",
       "      <td>0.956565</td>\n",
       "      <td>-0.333872</td>\n",
       "      <td>0.021822</td>\n",
       "      <td>-0.14241</td>\n",
       "      <td>-0.503494</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090786</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.211778</td>\n",
       "      <td>-1.587117</td>\n",
       "      <td>0.381667</td>\n",
       "      <td>2.33641</td>\n",
       "      <td>-0.953477</td>\n",
       "      <td>0.329873</td>\n",
       "      <td>-0.225695</td>\n",
       "      <td>-1.577533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-C5-A1BN-01</th>\n",
       "      <td>1</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.712492</td>\n",
       "      <td>-1.736526</td>\n",
       "      <td>-0.093635</td>\n",
       "      <td>-0.467472</td>\n",
       "      <td>-0.217678</td>\n",
       "      <td>-0.56891</td>\n",
       "      <td>-2.295594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.490214</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.023722</td>\n",
       "      <td>-1.149717</td>\n",
       "      <td>-2.535933</td>\n",
       "      <td>1.01021</td>\n",
       "      <td>0.157723</td>\n",
       "      <td>-0.763227</td>\n",
       "      <td>-1.524095</td>\n",
       "      <td>-1.707333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Cervical Cancer   Age  Gender  ARHGEF10L     HIF3A     RNF17  \\\n",
       "sampleID                                                                        \n",
       "TCGA-BI-A0VR-01                1  22.0     0.0  -0.352292 -2.344926  0.592565   \n",
       "TCGA-C5-A1BI-01                1  16.0     0.0   0.403208 -2.758326  0.462565   \n",
       "TCGA-C5-A1BJ-01                1  24.0     0.0  -0.314392 -4.098426  2.206165   \n",
       "TCGA-C5-A1BL-01                1  16.0     0.0  -0.455892 -1.485426  0.956565   \n",
       "TCGA-C5-A1BN-01                1  15.0     0.0  -1.712492 -1.736526 -0.093635   \n",
       "\n",
       "                    RNF10     RNF11    RNF13   GTF2IP1  ...   SLC7A10  \\\n",
       "sampleID                                                ...             \n",
       "TCGA-BI-A0VR-01 -0.256772  0.176622 -0.02901 -0.346094  ... -2.090786   \n",
       "TCGA-C5-A1BI-01  0.011628  0.168622 -0.05601 -0.784594  ... -2.090786   \n",
       "TCGA-C5-A1BJ-01 -0.411172  0.083522 -0.28401 -0.401694  ... -1.562186   \n",
       "TCGA-C5-A1BL-01 -0.333872  0.021822 -0.14241 -0.503494  ... -2.090786   \n",
       "TCGA-C5-A1BN-01 -0.467472 -0.217678 -0.56891 -2.295594  ...  0.490214   \n",
       "\n",
       "                  PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1     TULP3  \\\n",
       "sampleID                                                                     \n",
       "TCGA-BI-A0VR-01 -0.086682 -0.270678 -1.587117  0.725467  0.08141  0.103323   \n",
       "TCGA-C5-A1BI-01 -0.086682 -0.016978 -1.587117  0.482567  0.29501  0.262323   \n",
       "TCGA-C5-A1BJ-01 -0.086682 -0.748878 -0.682517  0.168067  2.51161  0.363123   \n",
       "TCGA-C5-A1BL-01 -0.086682 -0.211778 -1.587117  0.381667  2.33641 -0.953477   \n",
       "TCGA-C5-A1BN-01 -0.086682  0.023722 -1.149717 -2.535933  1.01021  0.157723   \n",
       "\n",
       "                    BCL6B     GSTK1      SELP  \n",
       "sampleID                                       \n",
       "TCGA-BI-A0VR-01 -0.890027 -0.151595  0.767467  \n",
       "TCGA-C5-A1BI-01 -0.657127 -0.189895  0.948267  \n",
       "TCGA-C5-A1BJ-01  0.377473 -0.686295  1.108467  \n",
       "TCGA-C5-A1BL-01  0.329873 -0.225695 -1.577533  \n",
       "TCGA-C5-A1BN-01 -0.763227 -1.524095 -1.707333  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.731303752Z",
     "start_time": "2024-02-27T00:23:16.717316474Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 88 samples.\n",
      "For the feature 'Cervical Cancer', the least common label is '0' with 1 occurrences. This represents 1.14% of the dataset.\n",
      "The distribution of the feature 'Cervical Cancer' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 16.0\n",
      "  50% (Median): 18.0\n",
      "  75%: 24.25\n",
      "Min: 11.0\n",
      "Max: 44.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 88 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.252387093Z",
     "start_time": "2024-02-27T00:23:16.722533204Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302271525Z",
     "start_time": "2024-02-27T00:23:30.295066253Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output2\\Jiayi\\Cervical-Cancer\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE137034', 'GSE172159', 'GSE192897']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Cervical-Cancer\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Cervical-Cancer\\\\GSE137034\\\\GSE137034_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Cervical-Cancer\\\\GSE137034\\\\GSE137034-GPL10558_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE137034\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Cervical-Cancer\\\\GSE172159\\\\GSE172159_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Cervical-Cancer\\\\GSE172159\\\\GSE172159_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE172159\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Cervical-Cancer\\\\GSE192897\\\\GSE192897_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Cervical-Cancer\\\\GSE192897\\\\GSE192897_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE192897\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Identifying molecular changes in early cervical cancer samples of patients that developed metastasis\"\n",
      "!Series_summary\t\"Cervical cancer is one of the most common cancers in women worldwide. Patients  diagnosed with early-stage cervical cancer have a good prognosis, however, 10-20%  suffer from local or distant recurrent disease after primary treatment. Treatment options for  recurrent cervical cancer are limited. Therefore, it is crucial to identify factors that can  predict patients with an increased risk of recurrence to optimize treatment to prevent the  recurrence of cervical cancer. We aimed to identify biomarkers in early-stage primary  cervical cancer which recurred after surgery. Formalin-Fixed, Paraffin-Embedded surgical  specimens of 34 patients with early-stage cervical cancer (FIGO 2009 stage 1B1) and 7  healthy controls were analyzed. Targeted gene expression profiling using the PanCancer  IO 360 panel of NanoString Technology was performed. The findings were confirmed by  performing immunohistochemistry stainings. Various genes, namely GLS, CD36, WNT5a,  HRAS, DDB2, PIK3R2, and CDH2 were found to be differentially highly expressed in  primary cervical cancer samples of patients who developed distant recurrence. In  addition, The relative infiltration score of CD8+ T cells, CD80+CD86+ macrophages,  CD163+MRC1+ macrophages, and FOXP3+IL2RA+ regulatory T cells were significantly  higher in this group of samples. In contrast, no significant differences in gene expression  and relative immune infiltration were found in samples of patients who developed local  recurrence. The infiltration of CD8 and FOXP3 cells were validated by  immunohistochemistry using all samples included in the study. We identified molecular  alterations in primary cervical cancer samples from patients who developed recurrent Q9  disease. These findings can be utilized towards developing a molecular signature for the  early detection of patients with a high risk to develop metastasis.\"\n",
      "!Series_overall_design\t\"RNA was extracted from FFPE samples (healthy n=7*2, CCLR n=14, CCDR n=8, CCNR n=12)\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM5768223</th>\n",
       "      <th>GSM5768224</th>\n",
       "      <th>GSM5768225</th>\n",
       "      <th>GSM5768226</th>\n",
       "      <th>GSM5768227</th>\n",
       "      <th>GSM5768228</th>\n",
       "      <th>GSM5768229</th>\n",
       "      <th>GSM5768230</th>\n",
       "      <th>GSM5768231</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5768261</th>\n",
       "      <th>GSM5768262</th>\n",
       "      <th>GSM5768263</th>\n",
       "      <th>GSM5768264</th>\n",
       "      <th>GSM5768265</th>\n",
       "      <th>GSM5768266</th>\n",
       "      <th>GSM5768267</th>\n",
       "      <th>GSM5768268</th>\n",
       "      <th>GSM5768269</th>\n",
       "      <th>GSM5768270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "      <td>tissue: cervix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>stage at last follow-up: Alive</td>\n",
       "      <td>stage at last follow-up: Death</td>\n",
       "      <td>stage at last follow-up: Alive</td>\n",
       "      <td>stage at last follow-up: Death</td>\n",
       "      <td>stage at last follow-up: Alive</td>\n",
       "      <td>stage at last follow-up: Death</td>\n",
       "      <td>stage at last follow-up: Death</td>\n",
       "      <td>stage at last follow-up: Death</td>\n",
       "      <td>stage at last follow-up: Alive</td>\n",
       "      <td>...</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "      <td>stage at last follow-up: n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 54</td>\n",
       "      <td>age: 37</td>\n",
       "      <td>age: 41</td>\n",
       "      <td>age: 60</td>\n",
       "      <td>age: 47</td>\n",
       "      <td>age: 59</td>\n",
       "      <td>age: 42</td>\n",
       "      <td>age: 41</td>\n",
       "      <td>age: 60</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>surgical approach: Open</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>histological subtype: adeno</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: squamous</td>\n",
       "      <td>histological subtype: adeno</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>lvsi: no</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>lvsi: yes</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tnm-stage: 1b1</td>\n",
       "      <td>tnm-stage: 1b1 N1</td>\n",
       "      <td>tnm-stage: 1b1 N1</td>\n",
       "      <td>tnm-stage: 1b1 N1</td>\n",
       "      <td>tnm-stage: 1b1 N1</td>\n",
       "      <td>tnm-stage: 1b1</td>\n",
       "      <td>tnm-stage: 1b1 N1</td>\n",
       "      <td>tnm-stage: 1b1 N1</td>\n",
       "      <td>tnm-stage: 1b1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>hpv: 18</td>\n",
       "      <td>hpv: 16</td>\n",
       "      <td>hpv: 16</td>\n",
       "      <td>hpv: 16</td>\n",
       "      <td>hpv: 18</td>\n",
       "      <td>hpv: 16</td>\n",
       "      <td>hpv: 16</td>\n",
       "      <td>hpv: 68 or 73</td>\n",
       "      <td>hpv: 18</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>adjuvant therapy: radiotherapy</td>\n",
       "      <td>adjuvant therapy: chemoradiation</td>\n",
       "      <td>adjuvant therapy: radiotherapy</td>\n",
       "      <td>adjuvant therapy: radiotherapy</td>\n",
       "      <td>adjuvant therapy: radiotherapy</td>\n",
       "      <td>adjuvant therapy: radiotherapy</td>\n",
       "      <td>adjuvant therapy: radiotherapy</td>\n",
       "      <td>adjuvant therapy: chemoradiation</td>\n",
       "      <td>adjuvant therapy: Follow up</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>time to recurrence: 90</td>\n",
       "      <td>time to recurrence: 89</td>\n",
       "      <td>time to recurrence: 22</td>\n",
       "      <td>time to recurrence: 39</td>\n",
       "      <td>time to recurrence: 45</td>\n",
       "      <td>time to recurrence: 23</td>\n",
       "      <td>time to recurrence: 13</td>\n",
       "      <td>time to recurrence: 10</td>\n",
       "      <td>time to recurrence: 57</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                      GSM5768223  \\\n",
       "0  !Sample_characteristics_ch1                  tissue: cervix   \n",
       "1  !Sample_characteristics_ch1  stage at last follow-up: Alive   \n",
       "2  !Sample_characteristics_ch1                         age: 54   \n",
       "3  !Sample_characteristics_ch1         surgical approach: Open   \n",
       "4  !Sample_characteristics_ch1     histological subtype: adeno   \n",
       "5  !Sample_characteristics_ch1                        lvsi: no   \n",
       "6  !Sample_characteristics_ch1                  tnm-stage: 1b1   \n",
       "7  !Sample_characteristics_ch1                         hpv: 18   \n",
       "8  !Sample_characteristics_ch1  adjuvant therapy: radiotherapy   \n",
       "9  !Sample_characteristics_ch1          time to recurrence: 90   \n",
       "\n",
       "                         GSM5768224                      GSM5768225  \\\n",
       "0                    tissue: cervix                  tissue: cervix   \n",
       "1    stage at last follow-up: Death  stage at last follow-up: Alive   \n",
       "2                           age: 37                         age: 41   \n",
       "3           surgical approach: Open         surgical approach: Open   \n",
       "4    histological subtype: squamous  histological subtype: squamous   \n",
       "5                         lvsi: yes                       lvsi: yes   \n",
       "6                 tnm-stage: 1b1 N1               tnm-stage: 1b1 N1   \n",
       "7                           hpv: 16                         hpv: 16   \n",
       "8  adjuvant therapy: chemoradiation  adjuvant therapy: radiotherapy   \n",
       "9            time to recurrence: 89          time to recurrence: 22   \n",
       "\n",
       "                       GSM5768226                      GSM5768227  \\\n",
       "0                  tissue: cervix                  tissue: cervix   \n",
       "1  stage at last follow-up: Death  stage at last follow-up: Alive   \n",
       "2                         age: 60                         age: 47   \n",
       "3         surgical approach: Open         surgical approach: Open   \n",
       "4  histological subtype: squamous  histological subtype: squamous   \n",
       "5                       lvsi: yes                       lvsi: yes   \n",
       "6               tnm-stage: 1b1 N1               tnm-stage: 1b1 N1   \n",
       "7                         hpv: 16                         hpv: 18   \n",
       "8  adjuvant therapy: radiotherapy  adjuvant therapy: radiotherapy   \n",
       "9          time to recurrence: 39          time to recurrence: 45   \n",
       "\n",
       "                       GSM5768228                      GSM5768229  \\\n",
       "0                  tissue: cervix                  tissue: cervix   \n",
       "1  stage at last follow-up: Death  stage at last follow-up: Death   \n",
       "2                         age: 59                         age: 42   \n",
       "3         surgical approach: Open         surgical approach: Open   \n",
       "4  histological subtype: squamous  histological subtype: squamous   \n",
       "5                       lvsi: yes                       lvsi: yes   \n",
       "6                  tnm-stage: 1b1               tnm-stage: 1b1 N1   \n",
       "7                         hpv: 16                         hpv: 16   \n",
       "8  adjuvant therapy: radiotherapy  adjuvant therapy: radiotherapy   \n",
       "9          time to recurrence: 23          time to recurrence: 13   \n",
       "\n",
       "                         GSM5768230                      GSM5768231  ...  \\\n",
       "0                    tissue: cervix                  tissue: cervix  ...   \n",
       "1    stage at last follow-up: Death  stage at last follow-up: Alive  ...   \n",
       "2                           age: 41                         age: 60  ...   \n",
       "3           surgical approach: Open         surgical approach: Open  ...   \n",
       "4    histological subtype: squamous     histological subtype: adeno  ...   \n",
       "5                         lvsi: yes                       lvsi: yes  ...   \n",
       "6                 tnm-stage: 1b1 N1                  tnm-stage: 1b1  ...   \n",
       "7                     hpv: 68 or 73                         hpv: 18  ...   \n",
       "8  adjuvant therapy: chemoradiation     adjuvant therapy: Follow up  ...   \n",
       "9            time to recurrence: 10          time to recurrence: 57  ...   \n",
       "\n",
       "                     GSM5768261                    GSM5768262  \\\n",
       "0                tissue: cervix                tissue: cervix   \n",
       "1  stage at last follow-up: n/a  stage at last follow-up: n/a   \n",
       "2                           NaN                           NaN   \n",
       "3                           NaN                           NaN   \n",
       "4                           NaN                           NaN   \n",
       "5                           NaN                           NaN   \n",
       "6                           NaN                           NaN   \n",
       "7                           NaN                           NaN   \n",
       "8                           NaN                           NaN   \n",
       "9                           NaN                           NaN   \n",
       "\n",
       "                     GSM5768263                    GSM5768264  \\\n",
       "0                tissue: cervix                tissue: cervix   \n",
       "1  stage at last follow-up: n/a  stage at last follow-up: n/a   \n",
       "2                           NaN                           NaN   \n",
       "3                           NaN                           NaN   \n",
       "4                           NaN                           NaN   \n",
       "5                           NaN                           NaN   \n",
       "6                           NaN                           NaN   \n",
       "7                           NaN                           NaN   \n",
       "8                           NaN                           NaN   \n",
       "9                           NaN                           NaN   \n",
       "\n",
       "                     GSM5768265                    GSM5768266  \\\n",
       "0                tissue: cervix                tissue: cervix   \n",
       "1  stage at last follow-up: n/a  stage at last follow-up: n/a   \n",
       "2                           NaN                           NaN   \n",
       "3                           NaN                           NaN   \n",
       "4                           NaN                           NaN   \n",
       "5                           NaN                           NaN   \n",
       "6                           NaN                           NaN   \n",
       "7                           NaN                           NaN   \n",
       "8                           NaN                           NaN   \n",
       "9                           NaN                           NaN   \n",
       "\n",
       "                     GSM5768267                    GSM5768268  \\\n",
       "0                tissue: cervix                tissue: cervix   \n",
       "1  stage at last follow-up: n/a  stage at last follow-up: n/a   \n",
       "2                           NaN                           NaN   \n",
       "3                           NaN                           NaN   \n",
       "4                           NaN                           NaN   \n",
       "5                           NaN                           NaN   \n",
       "6                           NaN                           NaN   \n",
       "7                           NaN                           NaN   \n",
       "8                           NaN                           NaN   \n",
       "9                           NaN                           NaN   \n",
       "\n",
       "                     GSM5768269                    GSM5768270  \n",
       "0                tissue: cervix                tissue: cervix  \n",
       "1  stage at last follow-up: n/a  stage at last follow-up: n/a  \n",
       "2                           NaN                           NaN  \n",
       "3                           NaN                           NaN  \n",
       "4                           NaN                           NaN  \n",
       "5                           NaN                           NaN  \n",
       "6                           NaN                           NaN  \n",
       "7                           NaN                           NaN  \n",
       "8                           NaN                           NaN  \n",
       "9                           NaN                           NaN  \n",
       "\n",
       "[10 rows x 49 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'stage at last follow-up: Alive',\n",
       "       'stage at last follow-up: Death', 'stage at last follow-up: n/a'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "trait_row = 1\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    if (trait == 'stage at last follow-up: Alive'):\n",
    "        return 1  \n",
    "    elif (trait == 'stage at last follow-up: Death'):\n",
    "        return 0  \n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5768223</th>\n",
       "      <th>GSM5768224</th>\n",
       "      <th>GSM5768225</th>\n",
       "      <th>GSM5768226</th>\n",
       "      <th>GSM5768227</th>\n",
       "      <th>GSM5768228</th>\n",
       "      <th>GSM5768229</th>\n",
       "      <th>GSM5768230</th>\n",
       "      <th>GSM5768231</th>\n",
       "      <th>GSM5768232</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5768261</th>\n",
       "      <th>GSM5768262</th>\n",
       "      <th>GSM5768263</th>\n",
       "      <th>GSM5768264</th>\n",
       "      <th>GSM5768265</th>\n",
       "      <th>GSM5768266</th>\n",
       "      <th>GSM5768267</th>\n",
       "      <th>GSM5768268</th>\n",
       "      <th>GSM5768269</th>\n",
       "      <th>GSM5768270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cervical Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM5768223  GSM5768224  GSM5768225  GSM5768226  GSM5768227  \\\n",
       "Cervical Cancer           1           0           1           0           1   \n",
       "\n",
       "                 GSM5768228  GSM5768229  GSM5768230  GSM5768231  GSM5768232  \\\n",
       "Cervical Cancer           0           0           0           1           1   \n",
       "\n",
       "                 ...  GSM5768261  GSM5768262  GSM5768263  GSM5768264  \\\n",
       "Cervical Cancer  ...        None        None        None        None   \n",
       "\n",
       "                 GSM5768265  GSM5768266  GSM5768267  GSM5768268  GSM5768269  \\\n",
       "Cervical Cancer        None        None        None        None        None   \n",
       "\n",
       "                 GSM5768270  \n",
       "Cervical Cancer        None  \n",
       "\n",
       "[1 rows x 48 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5768223</th>\n",
       "      <th>GSM5768224</th>\n",
       "      <th>GSM5768225</th>\n",
       "      <th>GSM5768226</th>\n",
       "      <th>GSM5768227</th>\n",
       "      <th>GSM5768228</th>\n",
       "      <th>GSM5768229</th>\n",
       "      <th>GSM5768230</th>\n",
       "      <th>GSM5768231</th>\n",
       "      <th>GSM5768232</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5768261</th>\n",
       "      <th>GSM5768262</th>\n",
       "      <th>GSM5768263</th>\n",
       "      <th>GSM5768264</th>\n",
       "      <th>GSM5768265</th>\n",
       "      <th>GSM5768266</th>\n",
       "      <th>GSM5768267</th>\n",
       "      <th>GSM5768268</th>\n",
       "      <th>GSM5768269</th>\n",
       "      <th>GSM5768270</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>13.640356</td>\n",
       "      <td>13.061804</td>\n",
       "      <td>13.092065</td>\n",
       "      <td>13.526674</td>\n",
       "      <td>13.281625</td>\n",
       "      <td>12.022688</td>\n",
       "      <td>12.160396</td>\n",
       "      <td>14.063438</td>\n",
       "      <td>12.870385</td>\n",
       "      <td>12.447663</td>\n",
       "      <td>...</td>\n",
       "      <td>14.199067</td>\n",
       "      <td>13.778855</td>\n",
       "      <td>13.403649</td>\n",
       "      <td>13.918247</td>\n",
       "      <td>12.970581</td>\n",
       "      <td>13.300940</td>\n",
       "      <td>14.335133</td>\n",
       "      <td>13.169667</td>\n",
       "      <td>13.206249</td>\n",
       "      <td>13.487752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1C</th>\n",
       "      <td>9.249743</td>\n",
       "      <td>8.866214</td>\n",
       "      <td>7.634368</td>\n",
       "      <td>5.429169</td>\n",
       "      <td>4.721611</td>\n",
       "      <td>5.490114</td>\n",
       "      <td>5.680905</td>\n",
       "      <td>7.724394</td>\n",
       "      <td>4.899190</td>\n",
       "      <td>7.749834</td>\n",
       "      <td>...</td>\n",
       "      <td>6.124174</td>\n",
       "      <td>8.942264</td>\n",
       "      <td>9.019673</td>\n",
       "      <td>8.514990</td>\n",
       "      <td>7.944484</td>\n",
       "      <td>8.114240</td>\n",
       "      <td>8.695080</td>\n",
       "      <td>8.326496</td>\n",
       "      <td>8.642206</td>\n",
       "      <td>8.776665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM12</th>\n",
       "      <td>9.820059</td>\n",
       "      <td>8.941878</td>\n",
       "      <td>9.542529</td>\n",
       "      <td>8.634718</td>\n",
       "      <td>10.839687</td>\n",
       "      <td>10.468202</td>\n",
       "      <td>8.154320</td>\n",
       "      <td>11.134893</td>\n",
       "      <td>8.203524</td>\n",
       "      <td>8.998410</td>\n",
       "      <td>...</td>\n",
       "      <td>10.002951</td>\n",
       "      <td>10.182578</td>\n",
       "      <td>8.043920</td>\n",
       "      <td>8.746316</td>\n",
       "      <td>9.804788</td>\n",
       "      <td>9.473643</td>\n",
       "      <td>8.889420</td>\n",
       "      <td>8.718376</td>\n",
       "      <td>8.646253</td>\n",
       "      <td>8.947777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADGRE1</th>\n",
       "      <td>8.676655</td>\n",
       "      <td>7.909935</td>\n",
       "      <td>3.895283</td>\n",
       "      <td>4.751097</td>\n",
       "      <td>5.265932</td>\n",
       "      <td>2.250648</td>\n",
       "      <td>2.680905</td>\n",
       "      <td>6.745980</td>\n",
       "      <td>4.247114</td>\n",
       "      <td>5.201397</td>\n",
       "      <td>...</td>\n",
       "      <td>6.956620</td>\n",
       "      <td>8.427690</td>\n",
       "      <td>8.959032</td>\n",
       "      <td>8.415718</td>\n",
       "      <td>7.203017</td>\n",
       "      <td>7.481022</td>\n",
       "      <td>8.873523</td>\n",
       "      <td>7.773343</td>\n",
       "      <td>8.126475</td>\n",
       "      <td>8.414095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADM</th>\n",
       "      <td>9.667758</td>\n",
       "      <td>11.137318</td>\n",
       "      <td>8.982745</td>\n",
       "      <td>11.248071</td>\n",
       "      <td>11.712660</td>\n",
       "      <td>11.277325</td>\n",
       "      <td>9.404931</td>\n",
       "      <td>9.541040</td>\n",
       "      <td>8.613019</td>\n",
       "      <td>11.628081</td>\n",
       "      <td>...</td>\n",
       "      <td>7.834316</td>\n",
       "      <td>8.787395</td>\n",
       "      <td>10.178298</td>\n",
       "      <td>11.320235</td>\n",
       "      <td>12.156122</td>\n",
       "      <td>11.706041</td>\n",
       "      <td>9.992378</td>\n",
       "      <td>11.550529</td>\n",
       "      <td>11.574802</td>\n",
       "      <td>11.810985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        GSM5768223  GSM5768224  GSM5768225  GSM5768226  GSM5768227  \\\n",
       "ID                                                                   \n",
       "A2M      13.640356   13.061804   13.092065   13.526674   13.281625   \n",
       "ACVR1C    9.249743    8.866214    7.634368    5.429169    4.721611   \n",
       "ADAM12    9.820059    8.941878    9.542529    8.634718   10.839687   \n",
       "ADGRE1    8.676655    7.909935    3.895283    4.751097    5.265932   \n",
       "ADM       9.667758   11.137318    8.982745   11.248071   11.712660   \n",
       "\n",
       "        GSM5768228  GSM5768229  GSM5768230  GSM5768231  GSM5768232  ...  \\\n",
       "ID                                                                  ...   \n",
       "A2M      12.022688   12.160396   14.063438   12.870385   12.447663  ...   \n",
       "ACVR1C    5.490114    5.680905    7.724394    4.899190    7.749834  ...   \n",
       "ADAM12   10.468202    8.154320   11.134893    8.203524    8.998410  ...   \n",
       "ADGRE1    2.250648    2.680905    6.745980    4.247114    5.201397  ...   \n",
       "ADM      11.277325    9.404931    9.541040    8.613019   11.628081  ...   \n",
       "\n",
       "        GSM5768261  GSM5768262  GSM5768263  GSM5768264  GSM5768265  \\\n",
       "ID                                                                   \n",
       "A2M      14.199067   13.778855   13.403649   13.918247   12.970581   \n",
       "ACVR1C    6.124174    8.942264    9.019673    8.514990    7.944484   \n",
       "ADAM12   10.002951   10.182578    8.043920    8.746316    9.804788   \n",
       "ADGRE1    6.956620    8.427690    8.959032    8.415718    7.203017   \n",
       "ADM       7.834316    8.787395   10.178298   11.320235   12.156122   \n",
       "\n",
       "        GSM5768266  GSM5768267  GSM5768268  GSM5768269  GSM5768270  \n",
       "ID                                                                  \n",
       "A2M      13.300940   14.335133   13.169667   13.206249   13.487752  \n",
       "ACVR1C    8.114240    8.695080    8.326496    8.642206    8.776665  \n",
       "ADAM12    9.473643    8.889420    8.718376    8.646253    8.947777  \n",
       "ADGRE1    7.481022    8.873523    7.773343    8.126475    8.414095  \n",
       "ADM      11.706041    9.992378   11.550529   11.574802   11.810985  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['A2M', 'ABCF1', 'ACVR1C', 'ADAM12', 'ADGRE1'], 'ORF': ['A2M', 'ABCF1', 'ACVR1C', 'ADAM12', 'ADGRE1'], 'SPOT_ID': [nan, nan, nan, nan, nan]}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'ORF', 'SPOT_ID'], dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'ORF'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5768223</th>\n",
       "      <th>GSM5768224</th>\n",
       "      <th>GSM5768225</th>\n",
       "      <th>GSM5768226</th>\n",
       "      <th>GSM5768227</th>\n",
       "      <th>GSM5768228</th>\n",
       "      <th>GSM5768229</th>\n",
       "      <th>GSM5768230</th>\n",
       "      <th>GSM5768231</th>\n",
       "      <th>GSM5768232</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5768261</th>\n",
       "      <th>GSM5768262</th>\n",
       "      <th>GSM5768263</th>\n",
       "      <th>GSM5768264</th>\n",
       "      <th>GSM5768265</th>\n",
       "      <th>GSM5768266</th>\n",
       "      <th>GSM5768267</th>\n",
       "      <th>GSM5768268</th>\n",
       "      <th>GSM5768269</th>\n",
       "      <th>GSM5768270</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>13.640356</td>\n",
       "      <td>13.061804</td>\n",
       "      <td>13.092065</td>\n",
       "      <td>13.526674</td>\n",
       "      <td>13.281625</td>\n",
       "      <td>12.022688</td>\n",
       "      <td>12.160396</td>\n",
       "      <td>14.063438</td>\n",
       "      <td>12.870385</td>\n",
       "      <td>12.447663</td>\n",
       "      <td>...</td>\n",
       "      <td>14.199067</td>\n",
       "      <td>13.778855</td>\n",
       "      <td>13.403649</td>\n",
       "      <td>13.918247</td>\n",
       "      <td>12.970581</td>\n",
       "      <td>13.300940</td>\n",
       "      <td>14.335133</td>\n",
       "      <td>13.169667</td>\n",
       "      <td>13.206249</td>\n",
       "      <td>13.487752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1C</th>\n",
       "      <td>9.249743</td>\n",
       "      <td>8.866214</td>\n",
       "      <td>7.634368</td>\n",
       "      <td>5.429169</td>\n",
       "      <td>4.721611</td>\n",
       "      <td>5.490114</td>\n",
       "      <td>5.680905</td>\n",
       "      <td>7.724394</td>\n",
       "      <td>4.899190</td>\n",
       "      <td>7.749834</td>\n",
       "      <td>...</td>\n",
       "      <td>6.124174</td>\n",
       "      <td>8.942264</td>\n",
       "      <td>9.019673</td>\n",
       "      <td>8.514990</td>\n",
       "      <td>7.944484</td>\n",
       "      <td>8.114240</td>\n",
       "      <td>8.695080</td>\n",
       "      <td>8.326496</td>\n",
       "      <td>8.642206</td>\n",
       "      <td>8.776665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM12</th>\n",
       "      <td>9.820059</td>\n",
       "      <td>8.941878</td>\n",
       "      <td>9.542529</td>\n",
       "      <td>8.634718</td>\n",
       "      <td>10.839687</td>\n",
       "      <td>10.468202</td>\n",
       "      <td>8.154320</td>\n",
       "      <td>11.134893</td>\n",
       "      <td>8.203524</td>\n",
       "      <td>8.998410</td>\n",
       "      <td>...</td>\n",
       "      <td>10.002951</td>\n",
       "      <td>10.182578</td>\n",
       "      <td>8.043920</td>\n",
       "      <td>8.746316</td>\n",
       "      <td>9.804788</td>\n",
       "      <td>9.473643</td>\n",
       "      <td>8.889420</td>\n",
       "      <td>8.718376</td>\n",
       "      <td>8.646253</td>\n",
       "      <td>8.947777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADGRE1</th>\n",
       "      <td>8.676655</td>\n",
       "      <td>7.909935</td>\n",
       "      <td>3.895283</td>\n",
       "      <td>4.751097</td>\n",
       "      <td>5.265932</td>\n",
       "      <td>2.250648</td>\n",
       "      <td>2.680905</td>\n",
       "      <td>6.745980</td>\n",
       "      <td>4.247114</td>\n",
       "      <td>5.201397</td>\n",
       "      <td>...</td>\n",
       "      <td>6.956620</td>\n",
       "      <td>8.427690</td>\n",
       "      <td>8.959032</td>\n",
       "      <td>8.415718</td>\n",
       "      <td>7.203017</td>\n",
       "      <td>7.481022</td>\n",
       "      <td>8.873523</td>\n",
       "      <td>7.773343</td>\n",
       "      <td>8.126475</td>\n",
       "      <td>8.414095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADM</th>\n",
       "      <td>9.667758</td>\n",
       "      <td>11.137318</td>\n",
       "      <td>8.982745</td>\n",
       "      <td>11.248071</td>\n",
       "      <td>11.712660</td>\n",
       "      <td>11.277325</td>\n",
       "      <td>9.404931</td>\n",
       "      <td>9.541040</td>\n",
       "      <td>8.613019</td>\n",
       "      <td>11.628081</td>\n",
       "      <td>...</td>\n",
       "      <td>7.834316</td>\n",
       "      <td>8.787395</td>\n",
       "      <td>10.178298</td>\n",
       "      <td>11.320235</td>\n",
       "      <td>12.156122</td>\n",
       "      <td>11.706041</td>\n",
       "      <td>9.992378</td>\n",
       "      <td>11.550529</td>\n",
       "      <td>11.574802</td>\n",
       "      <td>11.810985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WNT7B</th>\n",
       "      <td>8.658807</td>\n",
       "      <td>8.043519</td>\n",
       "      <td>9.147670</td>\n",
       "      <td>8.999025</td>\n",
       "      <td>9.509514</td>\n",
       "      <td>9.133291</td>\n",
       "      <td>9.090296</td>\n",
       "      <td>8.191057</td>\n",
       "      <td>8.370496</td>\n",
       "      <td>8.460669</td>\n",
       "      <td>...</td>\n",
       "      <td>5.920641</td>\n",
       "      <td>8.405438</td>\n",
       "      <td>8.379523</td>\n",
       "      <td>8.423601</td>\n",
       "      <td>8.362888</td>\n",
       "      <td>8.910706</td>\n",
       "      <td>8.522326</td>\n",
       "      <td>8.867926</td>\n",
       "      <td>9.199896</td>\n",
       "      <td>8.539053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZAP70</th>\n",
       "      <td>9.460163</td>\n",
       "      <td>8.936603</td>\n",
       "      <td>9.210832</td>\n",
       "      <td>10.624910</td>\n",
       "      <td>8.440429</td>\n",
       "      <td>7.342818</td>\n",
       "      <td>7.389250</td>\n",
       "      <td>9.454091</td>\n",
       "      <td>9.274756</td>\n",
       "      <td>8.696453</td>\n",
       "      <td>...</td>\n",
       "      <td>6.905994</td>\n",
       "      <td>8.409170</td>\n",
       "      <td>9.456379</td>\n",
       "      <td>8.363397</td>\n",
       "      <td>8.502225</td>\n",
       "      <td>8.022817</td>\n",
       "      <td>8.947523</td>\n",
       "      <td>8.108691</td>\n",
       "      <td>8.515201</td>\n",
       "      <td>9.277278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZC3H12A</th>\n",
       "      <td>10.003305</td>\n",
       "      <td>10.234015</td>\n",
       "      <td>8.041439</td>\n",
       "      <td>9.244232</td>\n",
       "      <td>10.005471</td>\n",
       "      <td>6.999586</td>\n",
       "      <td>8.833552</td>\n",
       "      <td>8.471473</td>\n",
       "      <td>10.103910</td>\n",
       "      <td>10.602276</td>\n",
       "      <td>...</td>\n",
       "      <td>7.679301</td>\n",
       "      <td>8.416607</td>\n",
       "      <td>10.285325</td>\n",
       "      <td>9.663853</td>\n",
       "      <td>9.051640</td>\n",
       "      <td>8.781196</td>\n",
       "      <td>8.932257</td>\n",
       "      <td>8.574907</td>\n",
       "      <td>8.773803</td>\n",
       "      <td>9.862241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZEB1</th>\n",
       "      <td>10.902000</td>\n",
       "      <td>9.792213</td>\n",
       "      <td>10.508259</td>\n",
       "      <td>10.218377</td>\n",
       "      <td>10.708401</td>\n",
       "      <td>9.691748</td>\n",
       "      <td>9.611642</td>\n",
       "      <td>11.804588</td>\n",
       "      <td>9.012143</td>\n",
       "      <td>9.443618</td>\n",
       "      <td>...</td>\n",
       "      <td>12.231047</td>\n",
       "      <td>11.113905</td>\n",
       "      <td>12.040051</td>\n",
       "      <td>11.664685</td>\n",
       "      <td>11.327784</td>\n",
       "      <td>11.652582</td>\n",
       "      <td>11.941817</td>\n",
       "      <td>11.315543</td>\n",
       "      <td>10.882435</td>\n",
       "      <td>11.585546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZEB2</th>\n",
       "      <td>10.909567</td>\n",
       "      <td>10.590361</td>\n",
       "      <td>10.907168</td>\n",
       "      <td>11.176691</td>\n",
       "      <td>10.756868</td>\n",
       "      <td>9.456848</td>\n",
       "      <td>9.085438</td>\n",
       "      <td>11.997287</td>\n",
       "      <td>9.787011</td>\n",
       "      <td>10.224874</td>\n",
       "      <td>...</td>\n",
       "      <td>12.041153</td>\n",
       "      <td>12.650214</td>\n",
       "      <td>12.798554</td>\n",
       "      <td>13.117069</td>\n",
       "      <td>13.141677</td>\n",
       "      <td>12.497489</td>\n",
       "      <td>12.796838</td>\n",
       "      <td>12.164697</td>\n",
       "      <td>12.511878</td>\n",
       "      <td>13.027122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>736 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM5768223  GSM5768224  GSM5768225  GSM5768226  GSM5768227  \\\n",
       "A2M       13.640356   13.061804   13.092065   13.526674   13.281625   \n",
       "ACVR1C     9.249743    8.866214    7.634368    5.429169    4.721611   \n",
       "ADAM12     9.820059    8.941878    9.542529    8.634718   10.839687   \n",
       "ADGRE1     8.676655    7.909935    3.895283    4.751097    5.265932   \n",
       "ADM        9.667758   11.137318    8.982745   11.248071   11.712660   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "WNT7B      8.658807    8.043519    9.147670    8.999025    9.509514   \n",
       "ZAP70      9.460163    8.936603    9.210832   10.624910    8.440429   \n",
       "ZC3H12A   10.003305   10.234015    8.041439    9.244232   10.005471   \n",
       "ZEB1      10.902000    9.792213   10.508259   10.218377   10.708401   \n",
       "ZEB2      10.909567   10.590361   10.907168   11.176691   10.756868   \n",
       "\n",
       "         GSM5768228  GSM5768229  GSM5768230  GSM5768231  GSM5768232  ...  \\\n",
       "A2M       12.022688   12.160396   14.063438   12.870385   12.447663  ...   \n",
       "ACVR1C     5.490114    5.680905    7.724394    4.899190    7.749834  ...   \n",
       "ADAM12    10.468202    8.154320   11.134893    8.203524    8.998410  ...   \n",
       "ADGRE1     2.250648    2.680905    6.745980    4.247114    5.201397  ...   \n",
       "ADM       11.277325    9.404931    9.541040    8.613019   11.628081  ...   \n",
       "...             ...         ...         ...         ...         ...  ...   \n",
       "WNT7B      9.133291    9.090296    8.191057    8.370496    8.460669  ...   \n",
       "ZAP70      7.342818    7.389250    9.454091    9.274756    8.696453  ...   \n",
       "ZC3H12A    6.999586    8.833552    8.471473   10.103910   10.602276  ...   \n",
       "ZEB1       9.691748    9.611642   11.804588    9.012143    9.443618  ...   \n",
       "ZEB2       9.456848    9.085438   11.997287    9.787011   10.224874  ...   \n",
       "\n",
       "         GSM5768261  GSM5768262  GSM5768263  GSM5768264  GSM5768265  \\\n",
       "A2M       14.199067   13.778855   13.403649   13.918247   12.970581   \n",
       "ACVR1C     6.124174    8.942264    9.019673    8.514990    7.944484   \n",
       "ADAM12    10.002951   10.182578    8.043920    8.746316    9.804788   \n",
       "ADGRE1     6.956620    8.427690    8.959032    8.415718    7.203017   \n",
       "ADM        7.834316    8.787395   10.178298   11.320235   12.156122   \n",
       "...             ...         ...         ...         ...         ...   \n",
       "WNT7B      5.920641    8.405438    8.379523    8.423601    8.362888   \n",
       "ZAP70      6.905994    8.409170    9.456379    8.363397    8.502225   \n",
       "ZC3H12A    7.679301    8.416607   10.285325    9.663853    9.051640   \n",
       "ZEB1      12.231047   11.113905   12.040051   11.664685   11.327784   \n",
       "ZEB2      12.041153   12.650214   12.798554   13.117069   13.141677   \n",
       "\n",
       "         GSM5768266  GSM5768267  GSM5768268  GSM5768269  GSM5768270  \n",
       "A2M       13.300940   14.335133   13.169667   13.206249   13.487752  \n",
       "ACVR1C     8.114240    8.695080    8.326496    8.642206    8.776665  \n",
       "ADAM12     9.473643    8.889420    8.718376    8.646253    8.947777  \n",
       "ADGRE1     7.481022    8.873523    7.773343    8.126475    8.414095  \n",
       "ADM       11.706041    9.992378   11.550529   11.574802   11.810985  \n",
       "...             ...         ...         ...         ...         ...  \n",
       "WNT7B      8.910706    8.522326    8.867926    9.199896    8.539053  \n",
       "ZAP70      8.022817    8.947523    8.108691    8.515201    9.277278  \n",
       "ZC3H12A    8.781196    8.932257    8.574907    8.773803    9.862241  \n",
       "ZEB1      11.652582   11.941817   11.315543   10.882435   11.585546  \n",
       "ZEB2      12.497489   12.796838   12.164697   12.511878   13.027122  \n",
       "\n",
       "[736 rows x 48 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:718: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  merged_data = pd.concat([clinical_df, genetic_df], axis=0).T.dropna()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cervical Cancer</th>\n",
       "      <th>A2M</th>\n",
       "      <th>ACVR1C</th>\n",
       "      <th>ADAM12</th>\n",
       "      <th>ADGRE1</th>\n",
       "      <th>ADM</th>\n",
       "      <th>ADORA2A</th>\n",
       "      <th>AKT1</th>\n",
       "      <th>ALDOA</th>\n",
       "      <th>ALDOC</th>\n",
       "      <th>...</th>\n",
       "      <th>WNT2B</th>\n",
       "      <th>WNT3A</th>\n",
       "      <th>WNT4</th>\n",
       "      <th>WNT5A</th>\n",
       "      <th>WNT5B</th>\n",
       "      <th>WNT7B</th>\n",
       "      <th>ZAP70</th>\n",
       "      <th>ZC3H12A</th>\n",
       "      <th>ZEB1</th>\n",
       "      <th>ZEB2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5768223</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.640356</td>\n",
       "      <td>9.249743</td>\n",
       "      <td>9.820059</td>\n",
       "      <td>8.676655</td>\n",
       "      <td>9.667758</td>\n",
       "      <td>9.176370</td>\n",
       "      <td>11.744508</td>\n",
       "      <td>13.710439</td>\n",
       "      <td>10.298586</td>\n",
       "      <td>...</td>\n",
       "      <td>9.851826</td>\n",
       "      <td>8.480627</td>\n",
       "      <td>9.652808</td>\n",
       "      <td>11.638460</td>\n",
       "      <td>9.435915</td>\n",
       "      <td>8.658807</td>\n",
       "      <td>9.460163</td>\n",
       "      <td>10.003305</td>\n",
       "      <td>10.902000</td>\n",
       "      <td>10.909567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768224</th>\n",
       "      <td>0.0</td>\n",
       "      <td>13.061804</td>\n",
       "      <td>8.866214</td>\n",
       "      <td>8.941878</td>\n",
       "      <td>7.909935</td>\n",
       "      <td>11.137318</td>\n",
       "      <td>8.218886</td>\n",
       "      <td>11.584470</td>\n",
       "      <td>14.557877</td>\n",
       "      <td>10.447473</td>\n",
       "      <td>...</td>\n",
       "      <td>9.957588</td>\n",
       "      <td>7.701827</td>\n",
       "      <td>9.274299</td>\n",
       "      <td>10.786361</td>\n",
       "      <td>8.849471</td>\n",
       "      <td>8.043519</td>\n",
       "      <td>8.936603</td>\n",
       "      <td>10.234015</td>\n",
       "      <td>9.792213</td>\n",
       "      <td>10.590361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768225</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.092065</td>\n",
       "      <td>7.634368</td>\n",
       "      <td>9.542529</td>\n",
       "      <td>3.895283</td>\n",
       "      <td>8.982745</td>\n",
       "      <td>6.957567</td>\n",
       "      <td>11.952797</td>\n",
       "      <td>14.312390</td>\n",
       "      <td>8.759104</td>\n",
       "      <td>...</td>\n",
       "      <td>10.915691</td>\n",
       "      <td>4.714710</td>\n",
       "      <td>8.816248</td>\n",
       "      <td>10.897932</td>\n",
       "      <td>8.611951</td>\n",
       "      <td>9.147670</td>\n",
       "      <td>9.210832</td>\n",
       "      <td>8.041439</td>\n",
       "      <td>10.508259</td>\n",
       "      <td>10.907168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768226</th>\n",
       "      <td>0.0</td>\n",
       "      <td>13.526674</td>\n",
       "      <td>5.429169</td>\n",
       "      <td>8.634718</td>\n",
       "      <td>4.751097</td>\n",
       "      <td>11.248071</td>\n",
       "      <td>8.374027</td>\n",
       "      <td>11.426348</td>\n",
       "      <td>13.737963</td>\n",
       "      <td>7.946445</td>\n",
       "      <td>...</td>\n",
       "      <td>9.023120</td>\n",
       "      <td>7.061437</td>\n",
       "      <td>8.482280</td>\n",
       "      <td>10.979916</td>\n",
       "      <td>8.638622</td>\n",
       "      <td>8.999025</td>\n",
       "      <td>10.624910</td>\n",
       "      <td>9.244232</td>\n",
       "      <td>10.218377</td>\n",
       "      <td>11.176691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768227</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.281625</td>\n",
       "      <td>4.721611</td>\n",
       "      <td>10.839687</td>\n",
       "      <td>5.265932</td>\n",
       "      <td>11.712660</td>\n",
       "      <td>7.384576</td>\n",
       "      <td>12.203747</td>\n",
       "      <td>14.486759</td>\n",
       "      <td>7.736561</td>\n",
       "      <td>...</td>\n",
       "      <td>6.982139</td>\n",
       "      <td>5.306574</td>\n",
       "      <td>5.384576</td>\n",
       "      <td>10.951231</td>\n",
       "      <td>7.476499</td>\n",
       "      <td>9.509514</td>\n",
       "      <td>8.440429</td>\n",
       "      <td>10.005471</td>\n",
       "      <td>10.708401</td>\n",
       "      <td>10.756868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768228</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.022688</td>\n",
       "      <td>5.490114</td>\n",
       "      <td>10.468202</td>\n",
       "      <td>2.250648</td>\n",
       "      <td>11.277325</td>\n",
       "      <td>6.270548</td>\n",
       "      <td>11.888279</td>\n",
       "      <td>14.249865</td>\n",
       "      <td>9.651528</td>\n",
       "      <td>...</td>\n",
       "      <td>10.222192</td>\n",
       "      <td>7.080723</td>\n",
       "      <td>6.168186</td>\n",
       "      <td>11.892900</td>\n",
       "      <td>8.574579</td>\n",
       "      <td>9.133291</td>\n",
       "      <td>7.342818</td>\n",
       "      <td>6.999586</td>\n",
       "      <td>9.691748</td>\n",
       "      <td>9.456848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768229</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.160396</td>\n",
       "      <td>5.680905</td>\n",
       "      <td>8.154320</td>\n",
       "      <td>2.680905</td>\n",
       "      <td>9.404931</td>\n",
       "      <td>6.412708</td>\n",
       "      <td>11.170466</td>\n",
       "      <td>14.282543</td>\n",
       "      <td>9.161262</td>\n",
       "      <td>...</td>\n",
       "      <td>6.768367</td>\n",
       "      <td>4.121477</td>\n",
       "      <td>5.083003</td>\n",
       "      <td>9.701686</td>\n",
       "      <td>7.299814</td>\n",
       "      <td>9.090296</td>\n",
       "      <td>7.389250</td>\n",
       "      <td>8.833552</td>\n",
       "      <td>9.611642</td>\n",
       "      <td>9.085438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768230</th>\n",
       "      <td>0.0</td>\n",
       "      <td>14.063438</td>\n",
       "      <td>7.724394</td>\n",
       "      <td>11.134893</td>\n",
       "      <td>6.745980</td>\n",
       "      <td>9.541040</td>\n",
       "      <td>7.855953</td>\n",
       "      <td>12.178212</td>\n",
       "      <td>13.359815</td>\n",
       "      <td>8.958134</td>\n",
       "      <td>...</td>\n",
       "      <td>9.036136</td>\n",
       "      <td>5.257694</td>\n",
       "      <td>7.306937</td>\n",
       "      <td>11.754437</td>\n",
       "      <td>8.676484</td>\n",
       "      <td>8.191057</td>\n",
       "      <td>9.454091</td>\n",
       "      <td>8.471473</td>\n",
       "      <td>11.804588</td>\n",
       "      <td>11.997287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768231</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.870385</td>\n",
       "      <td>4.899190</td>\n",
       "      <td>8.203524</td>\n",
       "      <td>4.247114</td>\n",
       "      <td>8.613019</td>\n",
       "      <td>7.306007</td>\n",
       "      <td>11.793273</td>\n",
       "      <td>14.014116</td>\n",
       "      <td>8.952828</td>\n",
       "      <td>...</td>\n",
       "      <td>7.229836</td>\n",
       "      <td>6.700286</td>\n",
       "      <td>6.843481</td>\n",
       "      <td>10.613852</td>\n",
       "      <td>7.140198</td>\n",
       "      <td>8.370496</td>\n",
       "      <td>9.274756</td>\n",
       "      <td>10.103910</td>\n",
       "      <td>9.012143</td>\n",
       "      <td>9.787011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768232</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.447663</td>\n",
       "      <td>7.749834</td>\n",
       "      <td>8.998410</td>\n",
       "      <td>5.201397</td>\n",
       "      <td>11.628081</td>\n",
       "      <td>7.059378</td>\n",
       "      <td>11.935090</td>\n",
       "      <td>14.839154</td>\n",
       "      <td>9.999707</td>\n",
       "      <td>...</td>\n",
       "      <td>8.833665</td>\n",
       "      <td>5.523325</td>\n",
       "      <td>8.677130</td>\n",
       "      <td>10.924546</td>\n",
       "      <td>7.271786</td>\n",
       "      <td>8.460669</td>\n",
       "      <td>8.696453</td>\n",
       "      <td>10.602276</td>\n",
       "      <td>9.443618</td>\n",
       "      <td>10.224874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768233</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.573983</td>\n",
       "      <td>5.457308</td>\n",
       "      <td>9.344834</td>\n",
       "      <td>5.483304</td>\n",
       "      <td>9.393625</td>\n",
       "      <td>6.914353</td>\n",
       "      <td>12.696581</td>\n",
       "      <td>14.399610</td>\n",
       "      <td>9.500907</td>\n",
       "      <td>...</td>\n",
       "      <td>6.533930</td>\n",
       "      <td>5.457308</td>\n",
       "      <td>6.390194</td>\n",
       "      <td>10.379852</td>\n",
       "      <td>5.979729</td>\n",
       "      <td>8.703855</td>\n",
       "      <td>7.758098</td>\n",
       "      <td>8.167802</td>\n",
       "      <td>9.647492</td>\n",
       "      <td>9.794890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768234</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.002261</td>\n",
       "      <td>6.606479</td>\n",
       "      <td>7.591832</td>\n",
       "      <td>4.299050</td>\n",
       "      <td>12.100597</td>\n",
       "      <td>6.907859</td>\n",
       "      <td>11.458921</td>\n",
       "      <td>14.924941</td>\n",
       "      <td>9.426271</td>\n",
       "      <td>...</td>\n",
       "      <td>8.390750</td>\n",
       "      <td>4.884013</td>\n",
       "      <td>7.325850</td>\n",
       "      <td>9.642458</td>\n",
       "      <td>6.126869</td>\n",
       "      <td>7.971475</td>\n",
       "      <td>7.960116</td>\n",
       "      <td>8.373727</td>\n",
       "      <td>7.663622</td>\n",
       "      <td>9.121780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768235</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.666646</td>\n",
       "      <td>5.428008</td>\n",
       "      <td>8.538062</td>\n",
       "      <td>3.770896</td>\n",
       "      <td>10.332551</td>\n",
       "      <td>7.346581</td>\n",
       "      <td>11.997123</td>\n",
       "      <td>14.222266</td>\n",
       "      <td>9.853455</td>\n",
       "      <td>...</td>\n",
       "      <td>7.299275</td>\n",
       "      <td>4.928437</td>\n",
       "      <td>6.940821</td>\n",
       "      <td>10.415862</td>\n",
       "      <td>7.670369</td>\n",
       "      <td>7.953099</td>\n",
       "      <td>9.098362</td>\n",
       "      <td>10.502998</td>\n",
       "      <td>8.558296</td>\n",
       "      <td>9.675936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768236</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.502261</td>\n",
       "      <td>6.719710</td>\n",
       "      <td>9.582206</td>\n",
       "      <td>5.845241</td>\n",
       "      <td>9.382675</td>\n",
       "      <td>7.985171</td>\n",
       "      <td>11.904535</td>\n",
       "      <td>14.121710</td>\n",
       "      <td>9.835187</td>\n",
       "      <td>...</td>\n",
       "      <td>10.192329</td>\n",
       "      <td>5.790793</td>\n",
       "      <td>7.516618</td>\n",
       "      <td>11.180466</td>\n",
       "      <td>7.299807</td>\n",
       "      <td>8.913720</td>\n",
       "      <td>9.761009</td>\n",
       "      <td>9.669669</td>\n",
       "      <td>10.597157</td>\n",
       "      <td>11.810597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768237</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.265909</td>\n",
       "      <td>8.270326</td>\n",
       "      <td>10.898415</td>\n",
       "      <td>6.689279</td>\n",
       "      <td>11.045399</td>\n",
       "      <td>7.823580</td>\n",
       "      <td>12.123161</td>\n",
       "      <td>13.908998</td>\n",
       "      <td>9.222497</td>\n",
       "      <td>...</td>\n",
       "      <td>10.481474</td>\n",
       "      <td>6.447204</td>\n",
       "      <td>8.703866</td>\n",
       "      <td>11.839851</td>\n",
       "      <td>8.050545</td>\n",
       "      <td>9.488231</td>\n",
       "      <td>9.533034</td>\n",
       "      <td>10.558139</td>\n",
       "      <td>10.572518</td>\n",
       "      <td>10.955608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768238</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.350253</td>\n",
       "      <td>7.537798</td>\n",
       "      <td>6.748217</td>\n",
       "      <td>5.788859</td>\n",
       "      <td>8.525825</td>\n",
       "      <td>6.838103</td>\n",
       "      <td>12.401728</td>\n",
       "      <td>13.637830</td>\n",
       "      <td>8.724843</td>\n",
       "      <td>...</td>\n",
       "      <td>8.534814</td>\n",
       "      <td>6.573131</td>\n",
       "      <td>7.778806</td>\n",
       "      <td>9.013397</td>\n",
       "      <td>7.028325</td>\n",
       "      <td>8.758486</td>\n",
       "      <td>7.945364</td>\n",
       "      <td>9.036787</td>\n",
       "      <td>10.038895</td>\n",
       "      <td>10.225340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768239</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.877204</td>\n",
       "      <td>7.249705</td>\n",
       "      <td>11.097191</td>\n",
       "      <td>7.174253</td>\n",
       "      <td>7.849902</td>\n",
       "      <td>8.203012</td>\n",
       "      <td>12.264276</td>\n",
       "      <td>13.138720</td>\n",
       "      <td>9.765656</td>\n",
       "      <td>...</td>\n",
       "      <td>9.478646</td>\n",
       "      <td>6.800574</td>\n",
       "      <td>9.193489</td>\n",
       "      <td>9.945188</td>\n",
       "      <td>7.825449</td>\n",
       "      <td>6.943769</td>\n",
       "      <td>9.245103</td>\n",
       "      <td>10.228880</td>\n",
       "      <td>12.003571</td>\n",
       "      <td>11.825834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768240</th>\n",
       "      <td>0.0</td>\n",
       "      <td>13.533846</td>\n",
       "      <td>8.498074</td>\n",
       "      <td>9.143874</td>\n",
       "      <td>4.481772</td>\n",
       "      <td>10.054386</td>\n",
       "      <td>7.011025</td>\n",
       "      <td>12.102825</td>\n",
       "      <td>14.074993</td>\n",
       "      <td>9.332953</td>\n",
       "      <td>...</td>\n",
       "      <td>9.820002</td>\n",
       "      <td>4.929231</td>\n",
       "      <td>7.141282</td>\n",
       "      <td>9.755356</td>\n",
       "      <td>6.976537</td>\n",
       "      <td>8.307743</td>\n",
       "      <td>9.616665</td>\n",
       "      <td>9.099156</td>\n",
       "      <td>10.672495</td>\n",
       "      <td>10.917916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768241</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.263227</td>\n",
       "      <td>6.699758</td>\n",
       "      <td>9.839310</td>\n",
       "      <td>3.699758</td>\n",
       "      <td>12.460597</td>\n",
       "      <td>7.042150</td>\n",
       "      <td>12.113688</td>\n",
       "      <td>14.028113</td>\n",
       "      <td>8.543455</td>\n",
       "      <td>...</td>\n",
       "      <td>10.031955</td>\n",
       "      <td>7.187024</td>\n",
       "      <td>8.364079</td>\n",
       "      <td>10.854576</td>\n",
       "      <td>7.439298</td>\n",
       "      <td>9.452736</td>\n",
       "      <td>8.335347</td>\n",
       "      <td>10.061075</td>\n",
       "      <td>10.956472</td>\n",
       "      <td>10.626266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768242</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.915387</td>\n",
       "      <td>5.391373</td>\n",
       "      <td>8.268362</td>\n",
       "      <td>4.900387</td>\n",
       "      <td>10.498288</td>\n",
       "      <td>6.938861</td>\n",
       "      <td>11.856783</td>\n",
       "      <td>13.701462</td>\n",
       "      <td>9.224263</td>\n",
       "      <td>...</td>\n",
       "      <td>10.870221</td>\n",
       "      <td>5.012861</td>\n",
       "      <td>8.218410</td>\n",
       "      <td>10.629043</td>\n",
       "      <td>7.656718</td>\n",
       "      <td>8.971704</td>\n",
       "      <td>9.196745</td>\n",
       "      <td>8.642218</td>\n",
       "      <td>9.752980</td>\n",
       "      <td>10.386162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768243</th>\n",
       "      <td>0.0</td>\n",
       "      <td>12.727511</td>\n",
       "      <td>4.857832</td>\n",
       "      <td>10.104353</td>\n",
       "      <td>4.857832</td>\n",
       "      <td>11.154833</td>\n",
       "      <td>8.211469</td>\n",
       "      <td>12.403583</td>\n",
       "      <td>14.815525</td>\n",
       "      <td>8.392963</td>\n",
       "      <td>...</td>\n",
       "      <td>5.562376</td>\n",
       "      <td>4.495262</td>\n",
       "      <td>5.312398</td>\n",
       "      <td>9.813751</td>\n",
       "      <td>6.424873</td>\n",
       "      <td>8.499549</td>\n",
       "      <td>7.594798</td>\n",
       "      <td>10.425999</td>\n",
       "      <td>9.379069</td>\n",
       "      <td>10.141863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768244</th>\n",
       "      <td>0.0</td>\n",
       "      <td>13.429768</td>\n",
       "      <td>5.484690</td>\n",
       "      <td>9.391581</td>\n",
       "      <td>5.806618</td>\n",
       "      <td>10.991493</td>\n",
       "      <td>8.361207</td>\n",
       "      <td>11.871091</td>\n",
       "      <td>14.409317</td>\n",
       "      <td>8.713509</td>\n",
       "      <td>...</td>\n",
       "      <td>6.415427</td>\n",
       "      <td>6.128546</td>\n",
       "      <td>6.157115</td>\n",
       "      <td>9.421328</td>\n",
       "      <td>6.674515</td>\n",
       "      <td>8.039279</td>\n",
       "      <td>9.760814</td>\n",
       "      <td>9.099400</td>\n",
       "      <td>9.767778</td>\n",
       "      <td>10.866233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768245</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.548004</td>\n",
       "      <td>5.858155</td>\n",
       "      <td>9.031486</td>\n",
       "      <td>4.004006</td>\n",
       "      <td>8.510580</td>\n",
       "      <td>6.391029</td>\n",
       "      <td>11.946746</td>\n",
       "      <td>12.847081</td>\n",
       "      <td>6.687270</td>\n",
       "      <td>...</td>\n",
       "      <td>6.110921</td>\n",
       "      <td>1.303566</td>\n",
       "      <td>6.712957</td>\n",
       "      <td>9.107697</td>\n",
       "      <td>5.303566</td>\n",
       "      <td>8.336989</td>\n",
       "      <td>7.503238</td>\n",
       "      <td>8.222429</td>\n",
       "      <td>11.235042</td>\n",
       "      <td>10.371672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768246</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.335030</td>\n",
       "      <td>5.133905</td>\n",
       "      <td>10.290409</td>\n",
       "      <td>4.902579</td>\n",
       "      <td>9.721092</td>\n",
       "      <td>6.902579</td>\n",
       "      <td>11.961630</td>\n",
       "      <td>13.516008</td>\n",
       "      <td>7.390244</td>\n",
       "      <td>...</td>\n",
       "      <td>5.186372</td>\n",
       "      <td>5.022873</td>\n",
       "      <td>5.700945</td>\n",
       "      <td>9.652813</td>\n",
       "      <td>6.838449</td>\n",
       "      <td>7.133905</td>\n",
       "      <td>8.497958</td>\n",
       "      <td>9.312708</td>\n",
       "      <td>9.176679</td>\n",
       "      <td>10.771335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768247</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.204504</td>\n",
       "      <td>5.421688</td>\n",
       "      <td>9.905887</td>\n",
       "      <td>3.351299</td>\n",
       "      <td>7.971885</td>\n",
       "      <td>7.116833</td>\n",
       "      <td>11.936512</td>\n",
       "      <td>12.779938</td>\n",
       "      <td>7.351299</td>\n",
       "      <td>...</td>\n",
       "      <td>6.059118</td>\n",
       "      <td>5.059118</td>\n",
       "      <td>6.720532</td>\n",
       "      <td>9.040598</td>\n",
       "      <td>7.238824</td>\n",
       "      <td>8.225768</td>\n",
       "      <td>8.447223</td>\n",
       "      <td>8.123888</td>\n",
       "      <td>10.758425</td>\n",
       "      <td>11.269764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768248</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.457925</td>\n",
       "      <td>5.877478</td>\n",
       "      <td>8.506161</td>\n",
       "      <td>5.555549</td>\n",
       "      <td>9.621639</td>\n",
       "      <td>7.035019</td>\n",
       "      <td>12.125113</td>\n",
       "      <td>14.079564</td>\n",
       "      <td>7.362904</td>\n",
       "      <td>...</td>\n",
       "      <td>7.203838</td>\n",
       "      <td>5.281374</td>\n",
       "      <td>6.877478</td>\n",
       "      <td>9.898137</td>\n",
       "      <td>6.121896</td>\n",
       "      <td>8.548597</td>\n",
       "      <td>6.527535</td>\n",
       "      <td>10.394269</td>\n",
       "      <td>9.907678</td>\n",
       "      <td>10.165722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768249</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.352970</td>\n",
       "      <td>7.425234</td>\n",
       "      <td>10.330769</td>\n",
       "      <td>8.975926</td>\n",
       "      <td>10.113115</td>\n",
       "      <td>8.170304</td>\n",
       "      <td>12.386296</td>\n",
       "      <td>14.143441</td>\n",
       "      <td>8.542273</td>\n",
       "      <td>...</td>\n",
       "      <td>10.326696</td>\n",
       "      <td>7.297854</td>\n",
       "      <td>7.954967</td>\n",
       "      <td>11.198071</td>\n",
       "      <td>8.672091</td>\n",
       "      <td>8.849134</td>\n",
       "      <td>9.092470</td>\n",
       "      <td>9.370884</td>\n",
       "      <td>10.275487</td>\n",
       "      <td>10.854803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768250</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.861578</td>\n",
       "      <td>8.121680</td>\n",
       "      <td>9.724426</td>\n",
       "      <td>7.087732</td>\n",
       "      <td>10.521964</td>\n",
       "      <td>8.417881</td>\n",
       "      <td>11.866842</td>\n",
       "      <td>12.984742</td>\n",
       "      <td>9.812279</td>\n",
       "      <td>...</td>\n",
       "      <td>10.001021</td>\n",
       "      <td>8.428769</td>\n",
       "      <td>8.563312</td>\n",
       "      <td>9.151564</td>\n",
       "      <td>8.635582</td>\n",
       "      <td>8.708877</td>\n",
       "      <td>9.231468</td>\n",
       "      <td>9.174387</td>\n",
       "      <td>11.562694</td>\n",
       "      <td>11.558982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768251</th>\n",
       "      <td>1.0</td>\n",
       "      <td>14.547546</td>\n",
       "      <td>7.234268</td>\n",
       "      <td>11.366058</td>\n",
       "      <td>6.746330</td>\n",
       "      <td>9.703928</td>\n",
       "      <td>8.627183</td>\n",
       "      <td>12.504472</td>\n",
       "      <td>13.529129</td>\n",
       "      <td>9.018135</td>\n",
       "      <td>...</td>\n",
       "      <td>9.411146</td>\n",
       "      <td>6.476241</td>\n",
       "      <td>7.973741</td>\n",
       "      <td>9.657905</td>\n",
       "      <td>8.518061</td>\n",
       "      <td>8.252964</td>\n",
       "      <td>9.283596</td>\n",
       "      <td>9.274475</td>\n",
       "      <td>12.023589</td>\n",
       "      <td>11.958159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768252</th>\n",
       "      <td>1.0</td>\n",
       "      <td>13.412800</td>\n",
       "      <td>5.576922</td>\n",
       "      <td>10.109579</td>\n",
       "      <td>4.254994</td>\n",
       "      <td>9.596462</td>\n",
       "      <td>7.254994</td>\n",
       "      <td>12.037843</td>\n",
       "      <td>13.030679</td>\n",
       "      <td>7.221827</td>\n",
       "      <td>...</td>\n",
       "      <td>6.482063</td>\n",
       "      <td>4.439418</td>\n",
       "      <td>6.883025</td>\n",
       "      <td>10.490790</td>\n",
       "      <td>5.602917</td>\n",
       "      <td>8.637913</td>\n",
       "      <td>7.596462</td>\n",
       "      <td>7.914503</td>\n",
       "      <td>10.843368</td>\n",
       "      <td>10.874380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768253</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.856029</td>\n",
       "      <td>5.769229</td>\n",
       "      <td>9.467934</td>\n",
       "      <td>5.195494</td>\n",
       "      <td>11.539480</td>\n",
       "      <td>7.063850</td>\n",
       "      <td>12.397903</td>\n",
       "      <td>14.102944</td>\n",
       "      <td>7.735014</td>\n",
       "      <td>...</td>\n",
       "      <td>8.608433</td>\n",
       "      <td>4.292791</td>\n",
       "      <td>5.126781</td>\n",
       "      <td>9.176497</td>\n",
       "      <td>6.017157</td>\n",
       "      <td>9.350430</td>\n",
       "      <td>8.372856</td>\n",
       "      <td>9.052318</td>\n",
       "      <td>9.335283</td>\n",
       "      <td>10.633416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768254</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.503058</td>\n",
       "      <td>6.096596</td>\n",
       "      <td>7.754315</td>\n",
       "      <td>5.145506</td>\n",
       "      <td>11.037087</td>\n",
       "      <td>7.896827</td>\n",
       "      <td>12.059592</td>\n",
       "      <td>14.021716</td>\n",
       "      <td>8.304705</td>\n",
       "      <td>...</td>\n",
       "      <td>9.058794</td>\n",
       "      <td>4.993503</td>\n",
       "      <td>6.347140</td>\n",
       "      <td>8.762177</td>\n",
       "      <td>7.145506</td>\n",
       "      <td>8.669068</td>\n",
       "      <td>10.105894</td>\n",
       "      <td>8.993503</td>\n",
       "      <td>9.215895</td>\n",
       "      <td>10.825455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768255</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.414499</td>\n",
       "      <td>5.136204</td>\n",
       "      <td>10.229314</td>\n",
       "      <td>4.136204</td>\n",
       "      <td>10.604353</td>\n",
       "      <td>7.348255</td>\n",
       "      <td>12.088072</td>\n",
       "      <td>13.896536</td>\n",
       "      <td>7.672257</td>\n",
       "      <td>...</td>\n",
       "      <td>6.655578</td>\n",
       "      <td>6.111957</td>\n",
       "      <td>5.929753</td>\n",
       "      <td>10.193655</td>\n",
       "      <td>6.206594</td>\n",
       "      <td>9.243334</td>\n",
       "      <td>7.663942</td>\n",
       "      <td>7.957234</td>\n",
       "      <td>10.253068</td>\n",
       "      <td>10.837569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5768256</th>\n",
       "      <td>1.0</td>\n",
       "      <td>11.843185</td>\n",
       "      <td>5.645568</td>\n",
       "      <td>6.711156</td>\n",
       "      <td>4.541231</td>\n",
       "      <td>7.317335</td>\n",
       "      <td>5.742865</td>\n",
       "      <td>10.825246</td>\n",
       "      <td>13.531264</td>\n",
       "      <td>7.252726</td>\n",
       "      <td>...</td>\n",
       "      <td>5.409128</td>\n",
       "      <td>3.974191</td>\n",
       "      <td>5.026658</td>\n",
       "      <td>7.026658</td>\n",
       "      <td>4.742865</td>\n",
       "      <td>9.498914</td>\n",
       "      <td>6.891728</td>\n",
       "      <td>9.318651</td>\n",
       "      <td>8.401697</td>\n",
       "      <td>9.133688</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34 rows × 737 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Cervical Cancer        A2M    ACVR1C     ADAM12    ADGRE1  \\\n",
       "GSM5768223              1.0  13.640356  9.249743   9.820059  8.676655   \n",
       "GSM5768224              0.0  13.061804  8.866214   8.941878  7.909935   \n",
       "GSM5768225              1.0  13.092065  7.634368   9.542529  3.895283   \n",
       "GSM5768226              0.0  13.526674  5.429169   8.634718  4.751097   \n",
       "GSM5768227              1.0  13.281625  4.721611  10.839687  5.265932   \n",
       "GSM5768228              0.0  12.022688  5.490114  10.468202  2.250648   \n",
       "GSM5768229              0.0  12.160396  5.680905   8.154320  2.680905   \n",
       "GSM5768230              0.0  14.063438  7.724394  11.134893  6.745980   \n",
       "GSM5768231              1.0  12.870385  4.899190   8.203524  4.247114   \n",
       "GSM5768232              1.0  12.447663  7.749834   8.998410  5.201397   \n",
       "GSM5768233              1.0  12.573983  5.457308   9.344834  5.483304   \n",
       "GSM5768234              1.0  12.002261  6.606479   7.591832  4.299050   \n",
       "GSM5768235              0.0  12.666646  5.428008   8.538062  3.770896   \n",
       "GSM5768236              1.0  13.502261  6.719710   9.582206  5.845241   \n",
       "GSM5768237              1.0  12.265909  8.270326  10.898415  6.689279   \n",
       "GSM5768238              1.0  12.350253  7.537798   6.748217  5.788859   \n",
       "GSM5768239              1.0  13.877204  7.249705  11.097191  7.174253   \n",
       "GSM5768240              0.0  13.533846  8.498074   9.143874  4.481772   \n",
       "GSM5768241              1.0  13.263227  6.699758   9.839310  3.699758   \n",
       "GSM5768242              0.0  12.915387  5.391373   8.268362  4.900387   \n",
       "GSM5768243              0.0  12.727511  4.857832  10.104353  4.857832   \n",
       "GSM5768244              0.0  13.429768  5.484690   9.391581  5.806618   \n",
       "GSM5768245              1.0  13.548004  5.858155   9.031486  4.004006   \n",
       "GSM5768246              1.0  12.335030  5.133905  10.290409  4.902579   \n",
       "GSM5768247              1.0  13.204504  5.421688   9.905887  3.351299   \n",
       "GSM5768248              1.0  13.457925  5.877478   8.506161  5.555549   \n",
       "GSM5768249              1.0  13.352970  7.425234  10.330769  8.975926   \n",
       "GSM5768250              1.0  13.861578  8.121680   9.724426  7.087732   \n",
       "GSM5768251              1.0  14.547546  7.234268  11.366058  6.746330   \n",
       "GSM5768252              1.0  13.412800  5.576922  10.109579  4.254994   \n",
       "GSM5768253              1.0  12.856029  5.769229   9.467934  5.195494   \n",
       "GSM5768254              1.0  12.503058  6.096596   7.754315  5.145506   \n",
       "GSM5768255              1.0  12.414499  5.136204  10.229314  4.136204   \n",
       "GSM5768256              1.0  11.843185  5.645568   6.711156  4.541231   \n",
       "\n",
       "                  ADM   ADORA2A       AKT1      ALDOA      ALDOC  ...  \\\n",
       "GSM5768223   9.667758  9.176370  11.744508  13.710439  10.298586  ...   \n",
       "GSM5768224  11.137318  8.218886  11.584470  14.557877  10.447473  ...   \n",
       "GSM5768225   8.982745  6.957567  11.952797  14.312390   8.759104  ...   \n",
       "GSM5768226  11.248071  8.374027  11.426348  13.737963   7.946445  ...   \n",
       "GSM5768227  11.712660  7.384576  12.203747  14.486759   7.736561  ...   \n",
       "GSM5768228  11.277325  6.270548  11.888279  14.249865   9.651528  ...   \n",
       "GSM5768229   9.404931  6.412708  11.170466  14.282543   9.161262  ...   \n",
       "GSM5768230   9.541040  7.855953  12.178212  13.359815   8.958134  ...   \n",
       "GSM5768231   8.613019  7.306007  11.793273  14.014116   8.952828  ...   \n",
       "GSM5768232  11.628081  7.059378  11.935090  14.839154   9.999707  ...   \n",
       "GSM5768233   9.393625  6.914353  12.696581  14.399610   9.500907  ...   \n",
       "GSM5768234  12.100597  6.907859  11.458921  14.924941   9.426271  ...   \n",
       "GSM5768235  10.332551  7.346581  11.997123  14.222266   9.853455  ...   \n",
       "GSM5768236   9.382675  7.985171  11.904535  14.121710   9.835187  ...   \n",
       "GSM5768237  11.045399  7.823580  12.123161  13.908998   9.222497  ...   \n",
       "GSM5768238   8.525825  6.838103  12.401728  13.637830   8.724843  ...   \n",
       "GSM5768239   7.849902  8.203012  12.264276  13.138720   9.765656  ...   \n",
       "GSM5768240  10.054386  7.011025  12.102825  14.074993   9.332953  ...   \n",
       "GSM5768241  12.460597  7.042150  12.113688  14.028113   8.543455  ...   \n",
       "GSM5768242  10.498288  6.938861  11.856783  13.701462   9.224263  ...   \n",
       "GSM5768243  11.154833  8.211469  12.403583  14.815525   8.392963  ...   \n",
       "GSM5768244  10.991493  8.361207  11.871091  14.409317   8.713509  ...   \n",
       "GSM5768245   8.510580  6.391029  11.946746  12.847081   6.687270  ...   \n",
       "GSM5768246   9.721092  6.902579  11.961630  13.516008   7.390244  ...   \n",
       "GSM5768247   7.971885  7.116833  11.936512  12.779938   7.351299  ...   \n",
       "GSM5768248   9.621639  7.035019  12.125113  14.079564   7.362904  ...   \n",
       "GSM5768249  10.113115  8.170304  12.386296  14.143441   8.542273  ...   \n",
       "GSM5768250  10.521964  8.417881  11.866842  12.984742   9.812279  ...   \n",
       "GSM5768251   9.703928  8.627183  12.504472  13.529129   9.018135  ...   \n",
       "GSM5768252   9.596462  7.254994  12.037843  13.030679   7.221827  ...   \n",
       "GSM5768253  11.539480  7.063850  12.397903  14.102944   7.735014  ...   \n",
       "GSM5768254  11.037087  7.896827  12.059592  14.021716   8.304705  ...   \n",
       "GSM5768255  10.604353  7.348255  12.088072  13.896536   7.672257  ...   \n",
       "GSM5768256   7.317335  5.742865  10.825246  13.531264   7.252726  ...   \n",
       "\n",
       "                WNT2B     WNT3A      WNT4      WNT5A     WNT5B     WNT7B  \\\n",
       "GSM5768223   9.851826  8.480627  9.652808  11.638460  9.435915  8.658807   \n",
       "GSM5768224   9.957588  7.701827  9.274299  10.786361  8.849471  8.043519   \n",
       "GSM5768225  10.915691  4.714710  8.816248  10.897932  8.611951  9.147670   \n",
       "GSM5768226   9.023120  7.061437  8.482280  10.979916  8.638622  8.999025   \n",
       "GSM5768227   6.982139  5.306574  5.384576  10.951231  7.476499  9.509514   \n",
       "GSM5768228  10.222192  7.080723  6.168186  11.892900  8.574579  9.133291   \n",
       "GSM5768229   6.768367  4.121477  5.083003   9.701686  7.299814  9.090296   \n",
       "GSM5768230   9.036136  5.257694  7.306937  11.754437  8.676484  8.191057   \n",
       "GSM5768231   7.229836  6.700286  6.843481  10.613852  7.140198  8.370496   \n",
       "GSM5768232   8.833665  5.523325  8.677130  10.924546  7.271786  8.460669   \n",
       "GSM5768233   6.533930  5.457308  6.390194  10.379852  5.979729  8.703855   \n",
       "GSM5768234   8.390750  4.884013  7.325850   9.642458  6.126869  7.971475   \n",
       "GSM5768235   7.299275  4.928437  6.940821  10.415862  7.670369  7.953099   \n",
       "GSM5768236  10.192329  5.790793  7.516618  11.180466  7.299807  8.913720   \n",
       "GSM5768237  10.481474  6.447204  8.703866  11.839851  8.050545  9.488231   \n",
       "GSM5768238   8.534814  6.573131  7.778806   9.013397  7.028325  8.758486   \n",
       "GSM5768239   9.478646  6.800574  9.193489   9.945188  7.825449  6.943769   \n",
       "GSM5768240   9.820002  4.929231  7.141282   9.755356  6.976537  8.307743   \n",
       "GSM5768241  10.031955  7.187024  8.364079  10.854576  7.439298  9.452736   \n",
       "GSM5768242  10.870221  5.012861  8.218410  10.629043  7.656718  8.971704   \n",
       "GSM5768243   5.562376  4.495262  5.312398   9.813751  6.424873  8.499549   \n",
       "GSM5768244   6.415427  6.128546  6.157115   9.421328  6.674515  8.039279   \n",
       "GSM5768245   6.110921  1.303566  6.712957   9.107697  5.303566  8.336989   \n",
       "GSM5768246   5.186372  5.022873  5.700945   9.652813  6.838449  7.133905   \n",
       "GSM5768247   6.059118  5.059118  6.720532   9.040598  7.238824  8.225768   \n",
       "GSM5768248   7.203838  5.281374  6.877478   9.898137  6.121896  8.548597   \n",
       "GSM5768249  10.326696  7.297854  7.954967  11.198071  8.672091  8.849134   \n",
       "GSM5768250  10.001021  8.428769  8.563312   9.151564  8.635582  8.708877   \n",
       "GSM5768251   9.411146  6.476241  7.973741   9.657905  8.518061  8.252964   \n",
       "GSM5768252   6.482063  4.439418  6.883025  10.490790  5.602917  8.637913   \n",
       "GSM5768253   8.608433  4.292791  5.126781   9.176497  6.017157  9.350430   \n",
       "GSM5768254   9.058794  4.993503  6.347140   8.762177  7.145506  8.669068   \n",
       "GSM5768255   6.655578  6.111957  5.929753  10.193655  6.206594  9.243334   \n",
       "GSM5768256   5.409128  3.974191  5.026658   7.026658  4.742865  9.498914   \n",
       "\n",
       "                ZAP70    ZC3H12A       ZEB1       ZEB2  \n",
       "GSM5768223   9.460163  10.003305  10.902000  10.909567  \n",
       "GSM5768224   8.936603  10.234015   9.792213  10.590361  \n",
       "GSM5768225   9.210832   8.041439  10.508259  10.907168  \n",
       "GSM5768226  10.624910   9.244232  10.218377  11.176691  \n",
       "GSM5768227   8.440429  10.005471  10.708401  10.756868  \n",
       "GSM5768228   7.342818   6.999586   9.691748   9.456848  \n",
       "GSM5768229   7.389250   8.833552   9.611642   9.085438  \n",
       "GSM5768230   9.454091   8.471473  11.804588  11.997287  \n",
       "GSM5768231   9.274756  10.103910   9.012143   9.787011  \n",
       "GSM5768232   8.696453  10.602276   9.443618  10.224874  \n",
       "GSM5768233   7.758098   8.167802   9.647492   9.794890  \n",
       "GSM5768234   7.960116   8.373727   7.663622   9.121780  \n",
       "GSM5768235   9.098362  10.502998   8.558296   9.675936  \n",
       "GSM5768236   9.761009   9.669669  10.597157  11.810597  \n",
       "GSM5768237   9.533034  10.558139  10.572518  10.955608  \n",
       "GSM5768238   7.945364   9.036787  10.038895  10.225340  \n",
       "GSM5768239   9.245103  10.228880  12.003571  11.825834  \n",
       "GSM5768240   9.616665   9.099156  10.672495  10.917916  \n",
       "GSM5768241   8.335347  10.061075  10.956472  10.626266  \n",
       "GSM5768242   9.196745   8.642218   9.752980  10.386162  \n",
       "GSM5768243   7.594798  10.425999   9.379069  10.141863  \n",
       "GSM5768244   9.760814   9.099400   9.767778  10.866233  \n",
       "GSM5768245   7.503238   8.222429  11.235042  10.371672  \n",
       "GSM5768246   8.497958   9.312708   9.176679  10.771335  \n",
       "GSM5768247   8.447223   8.123888  10.758425  11.269764  \n",
       "GSM5768248   6.527535  10.394269   9.907678  10.165722  \n",
       "GSM5768249   9.092470   9.370884  10.275487  10.854803  \n",
       "GSM5768250   9.231468   9.174387  11.562694  11.558982  \n",
       "GSM5768251   9.283596   9.274475  12.023589  11.958159  \n",
       "GSM5768252   7.596462   7.914503  10.843368  10.874380  \n",
       "GSM5768253   8.372856   9.052318   9.335283  10.633416  \n",
       "GSM5768254  10.105894   8.993503   9.215895  10.825455  \n",
       "GSM5768255   7.663942   7.957234  10.253068  10.837569  \n",
       "GSM5768256   6.891728   9.318651   8.401697   9.133688  \n",
       "\n",
       "[34 rows x 737 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 34 samples.\n",
      "For the feature 'Cervical Cancer', the least common label is '0.0' with 10 occurrences. This represents 29.41% of the dataset.\n",
      "The distribution of the feature 'Cervical Cancer' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
