{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = \"Creutzfeldt-Jakob Disease\"\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE62699', 'GSE87629']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Creutzfeldt-Jakob-Disease\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Creutzfeldt-Jakob-Disease\\\\GSE62699\\\\GSE62699_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Creutzfeldt-Jakob-Disease\\\\GSE62699\\\\GSE62699-GPL16384_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE62699\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Creutzfeldt-Jakob-Disease\\\\GSE87629\\\\GSE87629_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Creutzfeldt-Jakob-Disease\\\\GSE87629\\\\GSE87629-GPL10558_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE87629\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Genome-wide analysis of B and T cell gene expression during a six-week gluten challenge in patients with celiac disease\"\n",
      "!Series_summary\t\"Dietary gluten proteins (prolamins) from wheat, rye, and barley are the driving forces behind celiac disease, an organ-specific autoimmune disorder that targets both the small intestine and organs outside the gut. In the small intestine, gluten induces inflammation and a typical morphological change of villous atrophy and crypt hyperplasia. Gut lesions improve and heal when gluten is excluded from the diet and the disease relapses when patients consume gluten. Oral immune tolerance towards gluten may be kept for years or decades before breaking tolerance in genetically susceptible individuals. Celiac disease provides a unique opportunity to study autoimmunity and the transition in immune cells as gluten breaks oral tolerance. Seventy-three celiac disease patients on a long-term gluten-free diet ingested a known amount of gluten daily for six weeks. A peripheral blood sample and intestinal biopsies were taken before and six weeks after initiating the gluten challenge. Biopsy results were reported on a continuous numeric scale that measured the villus height to crypt depth ratio to quantify gluten-induced gut mucosal injury. Pooled B and T cells were isolated from whole blood, and RNA was analyzed by DNA microarray looking for changes in peripheral B- and T-cell gene expression that correlated with changes in villus height to crypt depth, as patients maintained or broke oral tolerance in the face of a gluten challenge.\"\n",
      "!Series_overall_design\t\"Whole blood sample were taken before and six weeks after a gluten challenge in 73 patients with celiac disease. B and T cells were purified from whole blood using anti-CD3 and anti-CD19 conjugated magnetic beads. Total RNA obtained from the purified pool of B and T cells was used for DNA microarray analysis using the Illumina platform.\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM2335776</th>\n",
       "      <th>GSM2335777</th>\n",
       "      <th>GSM2335778</th>\n",
       "      <th>GSM2335779</th>\n",
       "      <th>GSM2335780</th>\n",
       "      <th>GSM2335781</th>\n",
       "      <th>GSM2335782</th>\n",
       "      <th>GSM2335783</th>\n",
       "      <th>GSM2335784</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM2335877</th>\n",
       "      <th>GSM2335878</th>\n",
       "      <th>GSM2335879</th>\n",
       "      <th>GSM2335880</th>\n",
       "      <th>GSM2335881</th>\n",
       "      <th>GSM2335882</th>\n",
       "      <th>GSM2335883</th>\n",
       "      <th>GSM2335884</th>\n",
       "      <th>GSM2335885</th>\n",
       "      <th>GSM2335921</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>individual: celiac patient AA</td>\n",
       "      <td>individual: celiac patient AA</td>\n",
       "      <td>individual: celiac patient AB</td>\n",
       "      <td>individual: celiac patient AB</td>\n",
       "      <td>individual: celiac patient AC</td>\n",
       "      <td>individual: celiac patient AC</td>\n",
       "      <td>individual: celiac patient AD</td>\n",
       "      <td>individual: celiac patient AD</td>\n",
       "      <td>individual: celiac patient AE</td>\n",
       "      <td>...</td>\n",
       "      <td>individual: celiac patient DR</td>\n",
       "      <td>individual: celiac patient DW</td>\n",
       "      <td>individual: celiac patient DW</td>\n",
       "      <td>individual: celiac patient DX</td>\n",
       "      <td>individual: celiac patient DX</td>\n",
       "      <td>individual: celiac patient DY</td>\n",
       "      <td>individual: celiac patient DY</td>\n",
       "      <td>individual: celiac patient DZ</td>\n",
       "      <td>individual: celiac patient DZ</td>\n",
       "      <td>individual: celiac patient T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "      <td>disease state: biopsy confirmed celiac disease...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>...</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "      <td>treatment: 6 weeks gluten challenge</td>\n",
       "      <td>treatment: control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "      <td>tissue: peripheral whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "      <td>cell type: purified pool of B and T cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 108 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "\n",
       "                                          GSM2335776  \\\n",
       "0                      individual: celiac patient AA   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335777  \\\n",
       "0                      individual: celiac patient AA   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335778  \\\n",
       "0                      individual: celiac patient AB   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335779  \\\n",
       "0                      individual: celiac patient AB   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335780  \\\n",
       "0                      individual: celiac patient AC   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335781  \\\n",
       "0                      individual: celiac patient AC   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335782  \\\n",
       "0                      individual: celiac patient AD   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335783  \\\n",
       "0                      individual: celiac patient AD   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335784  ...  \\\n",
       "0                      individual: celiac patient AE  ...   \n",
       "1  disease state: biopsy confirmed celiac disease...  ...   \n",
       "2                                 treatment: control  ...   \n",
       "3                     tissue: peripheral whole blood  ...   \n",
       "4          cell type: purified pool of B and T cells  ...   \n",
       "\n",
       "                                          GSM2335877  \\\n",
       "0                      individual: celiac patient DR   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335878  \\\n",
       "0                      individual: celiac patient DW   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335879  \\\n",
       "0                      individual: celiac patient DW   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335880  \\\n",
       "0                      individual: celiac patient DX   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335881  \\\n",
       "0                      individual: celiac patient DX   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335882  \\\n",
       "0                      individual: celiac patient DY   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335883  \\\n",
       "0                      individual: celiac patient DY   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335884  \\\n",
       "0                      individual: celiac patient DZ   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                                 treatment: control   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335885  \\\n",
       "0                      individual: celiac patient DZ   \n",
       "1  disease state: biopsy confirmed celiac disease...   \n",
       "2                treatment: 6 weeks gluten challenge   \n",
       "3                     tissue: peripheral whole blood   \n",
       "4          cell type: purified pool of B and T cells   \n",
       "\n",
       "                                          GSM2335921  \n",
       "0                       individual: celiac patient T  \n",
       "1  disease state: biopsy confirmed celiac disease...  \n",
       "2                                 treatment: control  \n",
       "3                     tissue: peripheral whole blood  \n",
       "4          cell type: purified pool of B and T cells  \n",
       "\n",
       "[5 rows x 108 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'tissue: peripheral whole blood'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[3]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    if (trait == 'diagnosis: Control'):\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(':')[1])/12\n",
    "        age = round(age, 0)\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1531652</th>\n",
       "      <th>GSM1531653</th>\n",
       "      <th>GSM1531654</th>\n",
       "      <th>GSM1531655</th>\n",
       "      <th>GSM1531656</th>\n",
       "      <th>GSM1531657</th>\n",
       "      <th>GSM1531658</th>\n",
       "      <th>GSM1531659</th>\n",
       "      <th>GSM1531660</th>\n",
       "      <th>GSM1531661</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1531678</th>\n",
       "      <th>GSM1531679</th>\n",
       "      <th>GSM1531680</th>\n",
       "      <th>GSM1531681</th>\n",
       "      <th>GSM1531682</th>\n",
       "      <th>GSM1531683</th>\n",
       "      <th>GSM1531684</th>\n",
       "      <th>GSM1531685</th>\n",
       "      <th>GSM1531686</th>\n",
       "      <th>GSM1531687</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Creutzfeldt-Jakob Disease</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           GSM1531652  GSM1531653  GSM1531654  GSM1531655  \\\n",
       "Creutzfeldt-Jakob Disease           1           0           1           0   \n",
       "\n",
       "                           GSM1531656  GSM1531657  GSM1531658  GSM1531659  \\\n",
       "Creutzfeldt-Jakob Disease           1           0           1           0   \n",
       "\n",
       "                           GSM1531660  GSM1531661  ...  GSM1531678  \\\n",
       "Creutzfeldt-Jakob Disease           1           0  ...           1   \n",
       "\n",
       "                           GSM1531679  GSM1531680  GSM1531681  GSM1531682  \\\n",
       "Creutzfeldt-Jakob Disease           0           1           0           1   \n",
       "\n",
       "                           GSM1531683  GSM1531684  GSM1531685  GSM1531686  \\\n",
       "Creutzfeldt-Jakob Disease           0           1           0           1   \n",
       "\n",
       "                           GSM1531687  \n",
       "Creutzfeldt-Jakob Disease           0  \n",
       "\n",
       "[1 rows x 36 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1531652</th>\n",
       "      <th>GSM1531653</th>\n",
       "      <th>GSM1531654</th>\n",
       "      <th>GSM1531655</th>\n",
       "      <th>GSM1531656</th>\n",
       "      <th>GSM1531657</th>\n",
       "      <th>GSM1531658</th>\n",
       "      <th>GSM1531659</th>\n",
       "      <th>GSM1531660</th>\n",
       "      <th>GSM1531661</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1531678</th>\n",
       "      <th>GSM1531679</th>\n",
       "      <th>GSM1531680</th>\n",
       "      <th>GSM1531681</th>\n",
       "      <th>GSM1531682</th>\n",
       "      <th>GSM1531683</th>\n",
       "      <th>GSM1531684</th>\n",
       "      <th>GSM1531685</th>\n",
       "      <th>GSM1531686</th>\n",
       "      <th>GSM1531687</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14q0_st</th>\n",
       "      <td>7.68542</td>\n",
       "      <td>7.69338</td>\n",
       "      <td>8.05731</td>\n",
       "      <td>8.03301</td>\n",
       "      <td>7.41483</td>\n",
       "      <td>7.87933</td>\n",
       "      <td>7.61217</td>\n",
       "      <td>7.80203</td>\n",
       "      <td>7.81174</td>\n",
       "      <td>8.22454</td>\n",
       "      <td>...</td>\n",
       "      <td>7.97129</td>\n",
       "      <td>8.30401</td>\n",
       "      <td>7.80492</td>\n",
       "      <td>7.72724</td>\n",
       "      <td>7.80544</td>\n",
       "      <td>7.76272</td>\n",
       "      <td>7.97463</td>\n",
       "      <td>7.87084</td>\n",
       "      <td>7.74456</td>\n",
       "      <td>7.95184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14qI-1_st</th>\n",
       "      <td>5.27731</td>\n",
       "      <td>5.32070</td>\n",
       "      <td>4.65591</td>\n",
       "      <td>4.48387</td>\n",
       "      <td>4.84928</td>\n",
       "      <td>4.88192</td>\n",
       "      <td>4.69104</td>\n",
       "      <td>5.04664</td>\n",
       "      <td>4.85520</td>\n",
       "      <td>4.60723</td>\n",
       "      <td>...</td>\n",
       "      <td>4.84690</td>\n",
       "      <td>4.52504</td>\n",
       "      <td>4.72786</td>\n",
       "      <td>4.56759</td>\n",
       "      <td>5.10662</td>\n",
       "      <td>5.04491</td>\n",
       "      <td>4.90743</td>\n",
       "      <td>4.88524</td>\n",
       "      <td>5.29804</td>\n",
       "      <td>5.21946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14qI-1_x_st</th>\n",
       "      <td>4.55933</td>\n",
       "      <td>4.59639</td>\n",
       "      <td>4.69234</td>\n",
       "      <td>4.80472</td>\n",
       "      <td>4.89809</td>\n",
       "      <td>4.83344</td>\n",
       "      <td>4.81458</td>\n",
       "      <td>4.96099</td>\n",
       "      <td>4.79273</td>\n",
       "      <td>4.90661</td>\n",
       "      <td>...</td>\n",
       "      <td>4.67695</td>\n",
       "      <td>4.37150</td>\n",
       "      <td>4.52660</td>\n",
       "      <td>4.77767</td>\n",
       "      <td>4.70989</td>\n",
       "      <td>4.74682</td>\n",
       "      <td>4.71377</td>\n",
       "      <td>4.53930</td>\n",
       "      <td>4.71029</td>\n",
       "      <td>4.75213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14qI-2_st</th>\n",
       "      <td>4.67843</td>\n",
       "      <td>4.29837</td>\n",
       "      <td>4.42294</td>\n",
       "      <td>4.40797</td>\n",
       "      <td>4.53978</td>\n",
       "      <td>4.50486</td>\n",
       "      <td>4.60397</td>\n",
       "      <td>4.48906</td>\n",
       "      <td>4.51160</td>\n",
       "      <td>4.49264</td>\n",
       "      <td>...</td>\n",
       "      <td>4.39827</td>\n",
       "      <td>4.49887</td>\n",
       "      <td>4.47253</td>\n",
       "      <td>4.35049</td>\n",
       "      <td>4.45511</td>\n",
       "      <td>4.50782</td>\n",
       "      <td>4.98151</td>\n",
       "      <td>4.62313</td>\n",
       "      <td>4.46442</td>\n",
       "      <td>4.38716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14qI-3_x_st</th>\n",
       "      <td>6.65881</td>\n",
       "      <td>6.33247</td>\n",
       "      <td>6.53892</td>\n",
       "      <td>6.50822</td>\n",
       "      <td>6.67872</td>\n",
       "      <td>6.08335</td>\n",
       "      <td>6.12720</td>\n",
       "      <td>6.33515</td>\n",
       "      <td>6.37796</td>\n",
       "      <td>6.23352</td>\n",
       "      <td>...</td>\n",
       "      <td>6.67494</td>\n",
       "      <td>6.30227</td>\n",
       "      <td>6.28161</td>\n",
       "      <td>6.05855</td>\n",
       "      <td>6.15381</td>\n",
       "      <td>6.25624</td>\n",
       "      <td>5.84670</td>\n",
       "      <td>6.01996</td>\n",
       "      <td>6.07394</td>\n",
       "      <td>6.11179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             GSM1531652  GSM1531653  GSM1531654  GSM1531655  GSM1531656  \\\n",
       "ID                                                                        \n",
       "14q0_st         7.68542     7.69338     8.05731     8.03301     7.41483   \n",
       "14qI-1_st       5.27731     5.32070     4.65591     4.48387     4.84928   \n",
       "14qI-1_x_st     4.55933     4.59639     4.69234     4.80472     4.89809   \n",
       "14qI-2_st       4.67843     4.29837     4.42294     4.40797     4.53978   \n",
       "14qI-3_x_st     6.65881     6.33247     6.53892     6.50822     6.67872   \n",
       "\n",
       "             GSM1531657  GSM1531658  GSM1531659  GSM1531660  GSM1531661  ...  \\\n",
       "ID                                                                       ...   \n",
       "14q0_st         7.87933     7.61217     7.80203     7.81174     8.22454  ...   \n",
       "14qI-1_st       4.88192     4.69104     5.04664     4.85520     4.60723  ...   \n",
       "14qI-1_x_st     4.83344     4.81458     4.96099     4.79273     4.90661  ...   \n",
       "14qI-2_st       4.50486     4.60397     4.48906     4.51160     4.49264  ...   \n",
       "14qI-3_x_st     6.08335     6.12720     6.33515     6.37796     6.23352  ...   \n",
       "\n",
       "             GSM1531678  GSM1531679  GSM1531680  GSM1531681  GSM1531682  \\\n",
       "ID                                                                        \n",
       "14q0_st         7.97129     8.30401     7.80492     7.72724     7.80544   \n",
       "14qI-1_st       4.84690     4.52504     4.72786     4.56759     5.10662   \n",
       "14qI-1_x_st     4.67695     4.37150     4.52660     4.77767     4.70989   \n",
       "14qI-2_st       4.39827     4.49887     4.47253     4.35049     4.45511   \n",
       "14qI-3_x_st     6.67494     6.30227     6.28161     6.05855     6.15381   \n",
       "\n",
       "             GSM1531683  GSM1531684  GSM1531685  GSM1531686  GSM1531687  \n",
       "ID                                                                       \n",
       "14q0_st         7.76272     7.97463     7.87084     7.74456     7.95184  \n",
       "14qI-1_st       5.04491     4.90743     4.88524     5.29804     5.21946  \n",
       "14qI-1_x_st     4.74682     4.71377     4.53930     4.71029     4.75213  \n",
       "14qI-2_st       4.50782     4.98151     4.62313     4.46442     4.38716  \n",
       "14qI-3_x_st     6.25624     5.84670     6.01996     6.07394     6.11179  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1531652</th>\n",
       "      <th>GSM1531653</th>\n",
       "      <th>GSM1531654</th>\n",
       "      <th>GSM1531655</th>\n",
       "      <th>GSM1531656</th>\n",
       "      <th>GSM1531657</th>\n",
       "      <th>GSM1531658</th>\n",
       "      <th>GSM1531659</th>\n",
       "      <th>GSM1531660</th>\n",
       "      <th>GSM1531661</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1531678</th>\n",
       "      <th>GSM1531679</th>\n",
       "      <th>GSM1531680</th>\n",
       "      <th>GSM1531681</th>\n",
       "      <th>GSM1531682</th>\n",
       "      <th>GSM1531683</th>\n",
       "      <th>GSM1531684</th>\n",
       "      <th>GSM1531685</th>\n",
       "      <th>GSM1531686</th>\n",
       "      <th>GSM1531687</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [GSM1531652, GSM1531653, GSM1531654, GSM1531655, GSM1531656, GSM1531657, GSM1531658, GSM1531659, GSM1531660, GSM1531661, GSM1531662, GSM1531663, GSM1531664, GSM1531665, GSM1531666, GSM1531667, GSM1531668, GSM1531669, GSM1531670, GSM1531671, GSM1531672, GSM1531673, GSM1531674, GSM1531675, GSM1531676, GSM1531677, GSM1531678, GSM1531679, GSM1531680, GSM1531681, GSM1531682, GSM1531683, GSM1531684, GSM1531685, GSM1531686, GSM1531687]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 36 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Coronary artery disease</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3188436</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.303746</td>\n",
       "      <td>4.658333</td>\n",
       "      <td>4.127707</td>\n",
       "      <td>7.417824</td>\n",
       "      <td>6.324911</td>\n",
       "      <td>3.665135</td>\n",
       "      <td>5.750955</td>\n",
       "      <td>8.740156</td>\n",
       "      <td>...</td>\n",
       "      <td>8.267012</td>\n",
       "      <td>8.622072</td>\n",
       "      <td>6.511444</td>\n",
       "      <td>5.276233</td>\n",
       "      <td>2.765249</td>\n",
       "      <td>7.955894</td>\n",
       "      <td>3.021423</td>\n",
       "      <td>6.195087</td>\n",
       "      <td>7.666162</td>\n",
       "      <td>6.685938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188437</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.864738</td>\n",
       "      <td>5.006374</td>\n",
       "      <td>4.158720</td>\n",
       "      <td>7.130532</td>\n",
       "      <td>6.181629</td>\n",
       "      <td>3.338409</td>\n",
       "      <td>5.849932</td>\n",
       "      <td>8.358170</td>\n",
       "      <td>...</td>\n",
       "      <td>8.543126</td>\n",
       "      <td>8.324926</td>\n",
       "      <td>7.095449</td>\n",
       "      <td>5.176791</td>\n",
       "      <td>3.325044</td>\n",
       "      <td>7.976143</td>\n",
       "      <td>3.084119</td>\n",
       "      <td>6.139374</td>\n",
       "      <td>7.575070</td>\n",
       "      <td>6.360326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188438</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.625849</td>\n",
       "      <td>5.018469</td>\n",
       "      <td>4.170842</td>\n",
       "      <td>7.552365</td>\n",
       "      <td>6.117605</td>\n",
       "      <td>3.538927</td>\n",
       "      <td>6.097752</td>\n",
       "      <td>8.067759</td>\n",
       "      <td>...</td>\n",
       "      <td>8.579405</td>\n",
       "      <td>8.531927</td>\n",
       "      <td>6.416341</td>\n",
       "      <td>5.417870</td>\n",
       "      <td>3.438114</td>\n",
       "      <td>7.913347</td>\n",
       "      <td>3.071192</td>\n",
       "      <td>6.160805</td>\n",
       "      <td>7.734392</td>\n",
       "      <td>6.078101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188439</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.606136</td>\n",
       "      <td>5.531297</td>\n",
       "      <td>3.799715</td>\n",
       "      <td>7.700452</td>\n",
       "      <td>6.009400</td>\n",
       "      <td>3.692063</td>\n",
       "      <td>5.732469</td>\n",
       "      <td>8.310598</td>\n",
       "      <td>...</td>\n",
       "      <td>8.082617</td>\n",
       "      <td>8.230054</td>\n",
       "      <td>6.799698</td>\n",
       "      <td>5.506264</td>\n",
       "      <td>2.951082</td>\n",
       "      <td>7.974691</td>\n",
       "      <td>3.529228</td>\n",
       "      <td>6.137104</td>\n",
       "      <td>7.540814</td>\n",
       "      <td>6.454944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188440</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.753605</td>\n",
       "      <td>4.990348</td>\n",
       "      <td>4.097827</td>\n",
       "      <td>7.175898</td>\n",
       "      <td>6.141250</td>\n",
       "      <td>3.423539</td>\n",
       "      <td>5.830323</td>\n",
       "      <td>7.929667</td>\n",
       "      <td>...</td>\n",
       "      <td>8.484641</td>\n",
       "      <td>8.155123</td>\n",
       "      <td>6.836687</td>\n",
       "      <td>5.443652</td>\n",
       "      <td>3.425032</td>\n",
       "      <td>7.950576</td>\n",
       "      <td>3.190696</td>\n",
       "      <td>5.938086</td>\n",
       "      <td>7.502443</td>\n",
       "      <td>6.175240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188441</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.869394</td>\n",
       "      <td>5.579674</td>\n",
       "      <td>4.110009</td>\n",
       "      <td>7.198859</td>\n",
       "      <td>6.126123</td>\n",
       "      <td>3.479282</td>\n",
       "      <td>6.109326</td>\n",
       "      <td>8.421108</td>\n",
       "      <td>...</td>\n",
       "      <td>8.367078</td>\n",
       "      <td>8.174617</td>\n",
       "      <td>6.765051</td>\n",
       "      <td>5.656255</td>\n",
       "      <td>3.187053</td>\n",
       "      <td>8.021766</td>\n",
       "      <td>3.124840</td>\n",
       "      <td>5.975827</td>\n",
       "      <td>7.580190</td>\n",
       "      <td>6.206856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188442</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.978041</td>\n",
       "      <td>5.287926</td>\n",
       "      <td>4.116779</td>\n",
       "      <td>7.059702</td>\n",
       "      <td>6.100074</td>\n",
       "      <td>3.381674</td>\n",
       "      <td>5.655645</td>\n",
       "      <td>8.434831</td>\n",
       "      <td>...</td>\n",
       "      <td>8.549734</td>\n",
       "      <td>8.018434</td>\n",
       "      <td>6.915468</td>\n",
       "      <td>5.748913</td>\n",
       "      <td>3.353418</td>\n",
       "      <td>8.111779</td>\n",
       "      <td>3.085333</td>\n",
       "      <td>6.044921</td>\n",
       "      <td>7.583205</td>\n",
       "      <td>6.291254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188443</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.964216</td>\n",
       "      <td>4.991422</td>\n",
       "      <td>4.215373</td>\n",
       "      <td>7.533977</td>\n",
       "      <td>6.208076</td>\n",
       "      <td>3.679147</td>\n",
       "      <td>5.507761</td>\n",
       "      <td>8.412982</td>\n",
       "      <td>...</td>\n",
       "      <td>8.332112</td>\n",
       "      <td>8.471732</td>\n",
       "      <td>6.902637</td>\n",
       "      <td>5.675422</td>\n",
       "      <td>3.178719</td>\n",
       "      <td>7.987204</td>\n",
       "      <td>3.296036</td>\n",
       "      <td>6.076884</td>\n",
       "      <td>7.574004</td>\n",
       "      <td>6.533650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188444</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.224023</td>\n",
       "      <td>5.519005</td>\n",
       "      <td>3.882698</td>\n",
       "      <td>7.031556</td>\n",
       "      <td>6.092556</td>\n",
       "      <td>3.540713</td>\n",
       "      <td>5.990248</td>\n",
       "      <td>8.256269</td>\n",
       "      <td>...</td>\n",
       "      <td>8.839863</td>\n",
       "      <td>8.628709</td>\n",
       "      <td>6.812532</td>\n",
       "      <td>5.279751</td>\n",
       "      <td>2.934744</td>\n",
       "      <td>8.028546</td>\n",
       "      <td>2.894363</td>\n",
       "      <td>6.343047</td>\n",
       "      <td>7.621562</td>\n",
       "      <td>6.387210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188445</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.846065</td>\n",
       "      <td>6.423629</td>\n",
       "      <td>4.163691</td>\n",
       "      <td>7.272274</td>\n",
       "      <td>6.378582</td>\n",
       "      <td>3.549975</td>\n",
       "      <td>6.079201</td>\n",
       "      <td>8.475148</td>\n",
       "      <td>...</td>\n",
       "      <td>8.378584</td>\n",
       "      <td>8.070188</td>\n",
       "      <td>6.636148</td>\n",
       "      <td>5.507240</td>\n",
       "      <td>2.942788</td>\n",
       "      <td>7.933273</td>\n",
       "      <td>2.766647</td>\n",
       "      <td>5.914579</td>\n",
       "      <td>7.645101</td>\n",
       "      <td>6.051843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188446</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.287622</td>\n",
       "      <td>5.173498</td>\n",
       "      <td>4.104680</td>\n",
       "      <td>7.144834</td>\n",
       "      <td>6.221400</td>\n",
       "      <td>3.392771</td>\n",
       "      <td>5.699014</td>\n",
       "      <td>8.120977</td>\n",
       "      <td>...</td>\n",
       "      <td>8.678665</td>\n",
       "      <td>8.725884</td>\n",
       "      <td>6.528831</td>\n",
       "      <td>5.624415</td>\n",
       "      <td>2.946764</td>\n",
       "      <td>7.907700</td>\n",
       "      <td>2.993817</td>\n",
       "      <td>5.954905</td>\n",
       "      <td>7.621656</td>\n",
       "      <td>5.828861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188447</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.156010</td>\n",
       "      <td>4.271832</td>\n",
       "      <td>3.867312</td>\n",
       "      <td>6.947477</td>\n",
       "      <td>6.189114</td>\n",
       "      <td>3.293813</td>\n",
       "      <td>6.087747</td>\n",
       "      <td>8.196407</td>\n",
       "      <td>...</td>\n",
       "      <td>8.581336</td>\n",
       "      <td>8.356214</td>\n",
       "      <td>6.434403</td>\n",
       "      <td>5.729710</td>\n",
       "      <td>3.532458</td>\n",
       "      <td>8.069907</td>\n",
       "      <td>3.263687</td>\n",
       "      <td>6.221065</td>\n",
       "      <td>7.580450</td>\n",
       "      <td>6.350554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188448</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.358287</td>\n",
       "      <td>5.212619</td>\n",
       "      <td>3.977452</td>\n",
       "      <td>7.588719</td>\n",
       "      <td>6.130336</td>\n",
       "      <td>3.448889</td>\n",
       "      <td>5.900012</td>\n",
       "      <td>8.251974</td>\n",
       "      <td>...</td>\n",
       "      <td>8.576355</td>\n",
       "      <td>8.157130</td>\n",
       "      <td>6.688377</td>\n",
       "      <td>5.621063</td>\n",
       "      <td>3.106766</td>\n",
       "      <td>7.897594</td>\n",
       "      <td>3.114287</td>\n",
       "      <td>6.239084</td>\n",
       "      <td>7.604069</td>\n",
       "      <td>6.519019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188449</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.203532</td>\n",
       "      <td>4.895658</td>\n",
       "      <td>4.310683</td>\n",
       "      <td>7.515999</td>\n",
       "      <td>6.105656</td>\n",
       "      <td>3.618016</td>\n",
       "      <td>5.716508</td>\n",
       "      <td>8.234775</td>\n",
       "      <td>...</td>\n",
       "      <td>8.482769</td>\n",
       "      <td>8.087789</td>\n",
       "      <td>6.748323</td>\n",
       "      <td>5.565416</td>\n",
       "      <td>2.034895</td>\n",
       "      <td>8.076388</td>\n",
       "      <td>2.919525</td>\n",
       "      <td>6.158042</td>\n",
       "      <td>7.580968</td>\n",
       "      <td>6.462511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188450</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.792244</td>\n",
       "      <td>5.284337</td>\n",
       "      <td>3.998466</td>\n",
       "      <td>7.116458</td>\n",
       "      <td>6.221015</td>\n",
       "      <td>3.426820</td>\n",
       "      <td>5.957097</td>\n",
       "      <td>8.296558</td>\n",
       "      <td>...</td>\n",
       "      <td>8.244348</td>\n",
       "      <td>8.525339</td>\n",
       "      <td>6.844444</td>\n",
       "      <td>5.511094</td>\n",
       "      <td>3.315548</td>\n",
       "      <td>8.148355</td>\n",
       "      <td>3.241473</td>\n",
       "      <td>6.138881</td>\n",
       "      <td>7.593875</td>\n",
       "      <td>6.394661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188451</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.473386</td>\n",
       "      <td>5.147473</td>\n",
       "      <td>3.948080</td>\n",
       "      <td>7.511301</td>\n",
       "      <td>6.114161</td>\n",
       "      <td>3.586194</td>\n",
       "      <td>5.817617</td>\n",
       "      <td>8.257500</td>\n",
       "      <td>...</td>\n",
       "      <td>8.350960</td>\n",
       "      <td>8.398373</td>\n",
       "      <td>6.574859</td>\n",
       "      <td>5.489509</td>\n",
       "      <td>2.954867</td>\n",
       "      <td>7.812939</td>\n",
       "      <td>3.145337</td>\n",
       "      <td>6.227980</td>\n",
       "      <td>7.640666</td>\n",
       "      <td>6.453943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188452</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.045235</td>\n",
       "      <td>5.332369</td>\n",
       "      <td>3.675632</td>\n",
       "      <td>7.603001</td>\n",
       "      <td>6.061434</td>\n",
       "      <td>3.535629</td>\n",
       "      <td>5.659334</td>\n",
       "      <td>8.460221</td>\n",
       "      <td>...</td>\n",
       "      <td>8.228051</td>\n",
       "      <td>8.678777</td>\n",
       "      <td>6.443543</td>\n",
       "      <td>5.493094</td>\n",
       "      <td>2.808252</td>\n",
       "      <td>8.011013</td>\n",
       "      <td>3.180217</td>\n",
       "      <td>6.254845</td>\n",
       "      <td>7.654089</td>\n",
       "      <td>6.585905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188453</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.284654</td>\n",
       "      <td>5.732896</td>\n",
       "      <td>4.159351</td>\n",
       "      <td>7.137496</td>\n",
       "      <td>6.166971</td>\n",
       "      <td>3.442562</td>\n",
       "      <td>6.101288</td>\n",
       "      <td>8.535165</td>\n",
       "      <td>...</td>\n",
       "      <td>8.643783</td>\n",
       "      <td>8.474535</td>\n",
       "      <td>6.470823</td>\n",
       "      <td>5.413179</td>\n",
       "      <td>3.904765</td>\n",
       "      <td>7.984032</td>\n",
       "      <td>2.820515</td>\n",
       "      <td>6.068423</td>\n",
       "      <td>7.752699</td>\n",
       "      <td>6.145825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188454</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.722569</td>\n",
       "      <td>6.006830</td>\n",
       "      <td>3.795928</td>\n",
       "      <td>6.992641</td>\n",
       "      <td>6.070157</td>\n",
       "      <td>3.356160</td>\n",
       "      <td>5.921927</td>\n",
       "      <td>8.184894</td>\n",
       "      <td>...</td>\n",
       "      <td>8.646843</td>\n",
       "      <td>8.147372</td>\n",
       "      <td>6.752873</td>\n",
       "      <td>5.409370</td>\n",
       "      <td>4.186267</td>\n",
       "      <td>7.903906</td>\n",
       "      <td>2.796483</td>\n",
       "      <td>6.021599</td>\n",
       "      <td>7.611209</td>\n",
       "      <td>6.227735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188455</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.330736</td>\n",
       "      <td>4.503568</td>\n",
       "      <td>4.116442</td>\n",
       "      <td>7.129081</td>\n",
       "      <td>6.252498</td>\n",
       "      <td>3.223941</td>\n",
       "      <td>6.261953</td>\n",
       "      <td>8.091747</td>\n",
       "      <td>...</td>\n",
       "      <td>9.084296</td>\n",
       "      <td>7.936571</td>\n",
       "      <td>6.641239</td>\n",
       "      <td>5.313100</td>\n",
       "      <td>3.093092</td>\n",
       "      <td>7.944278</td>\n",
       "      <td>3.197185</td>\n",
       "      <td>6.256100</td>\n",
       "      <td>7.665015</td>\n",
       "      <td>6.287613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188456</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.921114</td>\n",
       "      <td>5.003951</td>\n",
       "      <td>3.743374</td>\n",
       "      <td>7.136506</td>\n",
       "      <td>6.170196</td>\n",
       "      <td>3.516566</td>\n",
       "      <td>5.865967</td>\n",
       "      <td>8.525821</td>\n",
       "      <td>...</td>\n",
       "      <td>8.487517</td>\n",
       "      <td>8.337527</td>\n",
       "      <td>6.807974</td>\n",
       "      <td>5.428088</td>\n",
       "      <td>2.997642</td>\n",
       "      <td>8.094447</td>\n",
       "      <td>2.703266</td>\n",
       "      <td>6.341254</td>\n",
       "      <td>7.618183</td>\n",
       "      <td>6.515540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188457</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.865641</td>\n",
       "      <td>6.105759</td>\n",
       "      <td>3.949613</td>\n",
       "      <td>7.527175</td>\n",
       "      <td>6.138481</td>\n",
       "      <td>3.578454</td>\n",
       "      <td>5.973020</td>\n",
       "      <td>8.496295</td>\n",
       "      <td>...</td>\n",
       "      <td>8.278341</td>\n",
       "      <td>8.222359</td>\n",
       "      <td>6.340251</td>\n",
       "      <td>5.601537</td>\n",
       "      <td>3.003468</td>\n",
       "      <td>8.019139</td>\n",
       "      <td>3.060047</td>\n",
       "      <td>6.187632</td>\n",
       "      <td>7.713456</td>\n",
       "      <td>6.297963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188458</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.111839</td>\n",
       "      <td>4.726825</td>\n",
       "      <td>4.325596</td>\n",
       "      <td>6.948296</td>\n",
       "      <td>6.125191</td>\n",
       "      <td>3.394086</td>\n",
       "      <td>6.050757</td>\n",
       "      <td>8.129618</td>\n",
       "      <td>...</td>\n",
       "      <td>8.840227</td>\n",
       "      <td>7.856005</td>\n",
       "      <td>6.839303</td>\n",
       "      <td>5.526579</td>\n",
       "      <td>2.773841</td>\n",
       "      <td>8.208267</td>\n",
       "      <td>2.942156</td>\n",
       "      <td>6.020802</td>\n",
       "      <td>7.935561</td>\n",
       "      <td>6.234126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188459</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.029210</td>\n",
       "      <td>5.617011</td>\n",
       "      <td>4.099885</td>\n",
       "      <td>7.509040</td>\n",
       "      <td>6.114697</td>\n",
       "      <td>3.650611</td>\n",
       "      <td>5.800308</td>\n",
       "      <td>8.583949</td>\n",
       "      <td>...</td>\n",
       "      <td>8.162037</td>\n",
       "      <td>8.501409</td>\n",
       "      <td>6.507173</td>\n",
       "      <td>5.365783</td>\n",
       "      <td>3.114520</td>\n",
       "      <td>8.052435</td>\n",
       "      <td>3.098041</td>\n",
       "      <td>5.865893</td>\n",
       "      <td>7.679000</td>\n",
       "      <td>6.382137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188460</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.051465</td>\n",
       "      <td>5.576377</td>\n",
       "      <td>3.900159</td>\n",
       "      <td>6.785121</td>\n",
       "      <td>6.296190</td>\n",
       "      <td>3.449241</td>\n",
       "      <td>5.665155</td>\n",
       "      <td>8.636230</td>\n",
       "      <td>...</td>\n",
       "      <td>8.716007</td>\n",
       "      <td>7.302711</td>\n",
       "      <td>6.544093</td>\n",
       "      <td>5.620752</td>\n",
       "      <td>3.223022</td>\n",
       "      <td>8.041334</td>\n",
       "      <td>3.263122</td>\n",
       "      <td>5.863600</td>\n",
       "      <td>7.936546</td>\n",
       "      <td>6.318112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188461</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.301221</td>\n",
       "      <td>5.501579</td>\n",
       "      <td>3.850512</td>\n",
       "      <td>7.159783</td>\n",
       "      <td>6.267744</td>\n",
       "      <td>3.537963</td>\n",
       "      <td>6.272080</td>\n",
       "      <td>8.534717</td>\n",
       "      <td>...</td>\n",
       "      <td>8.342549</td>\n",
       "      <td>8.503106</td>\n",
       "      <td>6.633021</td>\n",
       "      <td>5.420298</td>\n",
       "      <td>2.597714</td>\n",
       "      <td>8.070085</td>\n",
       "      <td>2.918022</td>\n",
       "      <td>6.145421</td>\n",
       "      <td>7.642127</td>\n",
       "      <td>6.492053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188462</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.220185</td>\n",
       "      <td>5.775806</td>\n",
       "      <td>4.083462</td>\n",
       "      <td>7.226576</td>\n",
       "      <td>6.364349</td>\n",
       "      <td>3.404627</td>\n",
       "      <td>6.175795</td>\n",
       "      <td>8.524912</td>\n",
       "      <td>...</td>\n",
       "      <td>8.138508</td>\n",
       "      <td>8.203245</td>\n",
       "      <td>6.784049</td>\n",
       "      <td>5.759685</td>\n",
       "      <td>3.848073</td>\n",
       "      <td>8.041505</td>\n",
       "      <td>3.013529</td>\n",
       "      <td>5.962614</td>\n",
       "      <td>7.820722</td>\n",
       "      <td>6.316718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188463</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.775258</td>\n",
       "      <td>5.748089</td>\n",
       "      <td>4.373176</td>\n",
       "      <td>7.047342</td>\n",
       "      <td>6.236709</td>\n",
       "      <td>3.550072</td>\n",
       "      <td>5.665327</td>\n",
       "      <td>8.825548</td>\n",
       "      <td>...</td>\n",
       "      <td>8.369407</td>\n",
       "      <td>8.218673</td>\n",
       "      <td>6.454139</td>\n",
       "      <td>5.443404</td>\n",
       "      <td>3.475552</td>\n",
       "      <td>8.211853</td>\n",
       "      <td>3.219150</td>\n",
       "      <td>5.813261</td>\n",
       "      <td>7.839113</td>\n",
       "      <td>6.431135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188464</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.712203</td>\n",
       "      <td>5.997114</td>\n",
       "      <td>4.226218</td>\n",
       "      <td>7.763064</td>\n",
       "      <td>6.102275</td>\n",
       "      <td>3.700852</td>\n",
       "      <td>6.081315</td>\n",
       "      <td>8.270876</td>\n",
       "      <td>...</td>\n",
       "      <td>8.361647</td>\n",
       "      <td>8.679320</td>\n",
       "      <td>6.507502</td>\n",
       "      <td>5.760880</td>\n",
       "      <td>3.430780</td>\n",
       "      <td>7.947054</td>\n",
       "      <td>3.171273</td>\n",
       "      <td>6.128784</td>\n",
       "      <td>7.663779</td>\n",
       "      <td>6.179315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188465</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.108376</td>\n",
       "      <td>4.846222</td>\n",
       "      <td>4.302424</td>\n",
       "      <td>6.974612</td>\n",
       "      <td>6.416129</td>\n",
       "      <td>3.273958</td>\n",
       "      <td>5.791213</td>\n",
       "      <td>8.328396</td>\n",
       "      <td>...</td>\n",
       "      <td>8.383100</td>\n",
       "      <td>7.967007</td>\n",
       "      <td>6.869358</td>\n",
       "      <td>6.155583</td>\n",
       "      <td>2.740770</td>\n",
       "      <td>8.183173</td>\n",
       "      <td>3.346364</td>\n",
       "      <td>6.064534</td>\n",
       "      <td>7.741242</td>\n",
       "      <td>6.435414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3188466</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.859673</td>\n",
       "      <td>5.033945</td>\n",
       "      <td>4.164832</td>\n",
       "      <td>7.586978</td>\n",
       "      <td>6.196540</td>\n",
       "      <td>3.743622</td>\n",
       "      <td>5.698246</td>\n",
       "      <td>8.412775</td>\n",
       "      <td>...</td>\n",
       "      <td>8.145227</td>\n",
       "      <td>8.517387</td>\n",
       "      <td>6.733282</td>\n",
       "      <td>5.546450</td>\n",
       "      <td>2.678560</td>\n",
       "      <td>8.090283</td>\n",
       "      <td>3.457542</td>\n",
       "      <td>5.960804</td>\n",
       "      <td>7.713401</td>\n",
       "      <td>6.401391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31 rows × 10058 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Coronary artery disease  Gender      A1CF       A2M     A2ML1  \\\n",
       "GSM3188436                      1.0     0.0  4.303746  4.658333  4.127707   \n",
       "GSM3188437                      1.0     0.0  3.864738  5.006374  4.158720   \n",
       "GSM3188438                      1.0     0.0  3.625849  5.018469  4.170842   \n",
       "GSM3188439                      1.0     0.0  2.606136  5.531297  3.799715   \n",
       "GSM3188440                      1.0     0.0  3.753605  4.990348  4.097827   \n",
       "GSM3188441                      1.0     0.0  3.869394  5.579674  4.110009   \n",
       "GSM3188442                      1.0     0.0  3.978041  5.287926  4.116779   \n",
       "GSM3188443                      1.0     0.0  3.964216  4.991422  4.215373   \n",
       "GSM3188444                      1.0     0.0  4.224023  5.519005  3.882698   \n",
       "GSM3188445                      1.0     0.0  3.846065  6.423629  4.163691   \n",
       "GSM3188446                      1.0     0.0  5.287622  5.173498  4.104680   \n",
       "GSM3188447                      1.0     0.0  3.156010  4.271832  3.867312   \n",
       "GSM3188448                      1.0     0.0  4.358287  5.212619  3.977452   \n",
       "GSM3188449                      1.0     0.0  3.203532  4.895658  4.310683   \n",
       "GSM3188450                      1.0     0.0  3.792244  5.284337  3.998466   \n",
       "GSM3188451                      1.0     0.0  4.473386  5.147473  3.948080   \n",
       "GSM3188452                      1.0     0.0  4.045235  5.332369  3.675632   \n",
       "GSM3188453                      1.0     0.0  4.284654  5.732896  4.159351   \n",
       "GSM3188454                      1.0     0.0  3.722569  6.006830  3.795928   \n",
       "GSM3188455                      1.0     0.0  3.330736  4.503568  4.116442   \n",
       "GSM3188456                      1.0     0.0  2.921114  5.003951  3.743374   \n",
       "GSM3188457                      1.0     0.0  2.865641  6.105759  3.949613   \n",
       "GSM3188458                      1.0     0.0  4.111839  4.726825  4.325596   \n",
       "GSM3188459                      1.0     0.0  4.029210  5.617011  4.099885   \n",
       "GSM3188460                      0.0     0.0  4.051465  5.576377  3.900159   \n",
       "GSM3188461                      0.0     0.0  3.301221  5.501579  3.850512   \n",
       "GSM3188462                      0.0     0.0  4.220185  5.775806  4.083462   \n",
       "GSM3188463                      0.0     0.0  3.775258  5.748089  4.373176   \n",
       "GSM3188464                      0.0     0.0  3.712203  5.997114  4.226218   \n",
       "GSM3188465                      0.0     0.0  4.108376  4.846222  4.302424   \n",
       "GSM3188466                      0.0     0.0  3.859673  5.033945  4.164832   \n",
       "\n",
       "                AAAS      AACS     AADAT     AAGAB      AAK1  ...    ZSWIM6  \\\n",
       "GSM3188436  7.417824  6.324911  3.665135  5.750955  8.740156  ...  8.267012   \n",
       "GSM3188437  7.130532  6.181629  3.338409  5.849932  8.358170  ...  8.543126   \n",
       "GSM3188438  7.552365  6.117605  3.538927  6.097752  8.067759  ...  8.579405   \n",
       "GSM3188439  7.700452  6.009400  3.692063  5.732469  8.310598  ...  8.082617   \n",
       "GSM3188440  7.175898  6.141250  3.423539  5.830323  7.929667  ...  8.484641   \n",
       "GSM3188441  7.198859  6.126123  3.479282  6.109326  8.421108  ...  8.367078   \n",
       "GSM3188442  7.059702  6.100074  3.381674  5.655645  8.434831  ...  8.549734   \n",
       "GSM3188443  7.533977  6.208076  3.679147  5.507761  8.412982  ...  8.332112   \n",
       "GSM3188444  7.031556  6.092556  3.540713  5.990248  8.256269  ...  8.839863   \n",
       "GSM3188445  7.272274  6.378582  3.549975  6.079201  8.475148  ...  8.378584   \n",
       "GSM3188446  7.144834  6.221400  3.392771  5.699014  8.120977  ...  8.678665   \n",
       "GSM3188447  6.947477  6.189114  3.293813  6.087747  8.196407  ...  8.581336   \n",
       "GSM3188448  7.588719  6.130336  3.448889  5.900012  8.251974  ...  8.576355   \n",
       "GSM3188449  7.515999  6.105656  3.618016  5.716508  8.234775  ...  8.482769   \n",
       "GSM3188450  7.116458  6.221015  3.426820  5.957097  8.296558  ...  8.244348   \n",
       "GSM3188451  7.511301  6.114161  3.586194  5.817617  8.257500  ...  8.350960   \n",
       "GSM3188452  7.603001  6.061434  3.535629  5.659334  8.460221  ...  8.228051   \n",
       "GSM3188453  7.137496  6.166971  3.442562  6.101288  8.535165  ...  8.643783   \n",
       "GSM3188454  6.992641  6.070157  3.356160  5.921927  8.184894  ...  8.646843   \n",
       "GSM3188455  7.129081  6.252498  3.223941  6.261953  8.091747  ...  9.084296   \n",
       "GSM3188456  7.136506  6.170196  3.516566  5.865967  8.525821  ...  8.487517   \n",
       "GSM3188457  7.527175  6.138481  3.578454  5.973020  8.496295  ...  8.278341   \n",
       "GSM3188458  6.948296  6.125191  3.394086  6.050757  8.129618  ...  8.840227   \n",
       "GSM3188459  7.509040  6.114697  3.650611  5.800308  8.583949  ...  8.162037   \n",
       "GSM3188460  6.785121  6.296190  3.449241  5.665155  8.636230  ...  8.716007   \n",
       "GSM3188461  7.159783  6.267744  3.537963  6.272080  8.534717  ...  8.342549   \n",
       "GSM3188462  7.226576  6.364349  3.404627  6.175795  8.524912  ...  8.138508   \n",
       "GSM3188463  7.047342  6.236709  3.550072  5.665327  8.825548  ...  8.369407   \n",
       "GSM3188464  7.763064  6.102275  3.700852  6.081315  8.270876  ...  8.361647   \n",
       "GSM3188465  6.974612  6.416129  3.273958  5.791213  8.328396  ...  8.383100   \n",
       "GSM3188466  7.586978  6.196540  3.743622  5.698246  8.412775  ...  8.145227   \n",
       "\n",
       "              ZSWIM7      ZW10    ZWILCH     ZWINT      ZXDC    ZYG11A  \\\n",
       "GSM3188436  8.622072  6.511444  5.276233  2.765249  7.955894  3.021423   \n",
       "GSM3188437  8.324926  7.095449  5.176791  3.325044  7.976143  3.084119   \n",
       "GSM3188438  8.531927  6.416341  5.417870  3.438114  7.913347  3.071192   \n",
       "GSM3188439  8.230054  6.799698  5.506264  2.951082  7.974691  3.529228   \n",
       "GSM3188440  8.155123  6.836687  5.443652  3.425032  7.950576  3.190696   \n",
       "GSM3188441  8.174617  6.765051  5.656255  3.187053  8.021766  3.124840   \n",
       "GSM3188442  8.018434  6.915468  5.748913  3.353418  8.111779  3.085333   \n",
       "GSM3188443  8.471732  6.902637  5.675422  3.178719  7.987204  3.296036   \n",
       "GSM3188444  8.628709  6.812532  5.279751  2.934744  8.028546  2.894363   \n",
       "GSM3188445  8.070188  6.636148  5.507240  2.942788  7.933273  2.766647   \n",
       "GSM3188446  8.725884  6.528831  5.624415  2.946764  7.907700  2.993817   \n",
       "GSM3188447  8.356214  6.434403  5.729710  3.532458  8.069907  3.263687   \n",
       "GSM3188448  8.157130  6.688377  5.621063  3.106766  7.897594  3.114287   \n",
       "GSM3188449  8.087789  6.748323  5.565416  2.034895  8.076388  2.919525   \n",
       "GSM3188450  8.525339  6.844444  5.511094  3.315548  8.148355  3.241473   \n",
       "GSM3188451  8.398373  6.574859  5.489509  2.954867  7.812939  3.145337   \n",
       "GSM3188452  8.678777  6.443543  5.493094  2.808252  8.011013  3.180217   \n",
       "GSM3188453  8.474535  6.470823  5.413179  3.904765  7.984032  2.820515   \n",
       "GSM3188454  8.147372  6.752873  5.409370  4.186267  7.903906  2.796483   \n",
       "GSM3188455  7.936571  6.641239  5.313100  3.093092  7.944278  3.197185   \n",
       "GSM3188456  8.337527  6.807974  5.428088  2.997642  8.094447  2.703266   \n",
       "GSM3188457  8.222359  6.340251  5.601537  3.003468  8.019139  3.060047   \n",
       "GSM3188458  7.856005  6.839303  5.526579  2.773841  8.208267  2.942156   \n",
       "GSM3188459  8.501409  6.507173  5.365783  3.114520  8.052435  3.098041   \n",
       "GSM3188460  7.302711  6.544093  5.620752  3.223022  8.041334  3.263122   \n",
       "GSM3188461  8.503106  6.633021  5.420298  2.597714  8.070085  2.918022   \n",
       "GSM3188462  8.203245  6.784049  5.759685  3.848073  8.041505  3.013529   \n",
       "GSM3188463  8.218673  6.454139  5.443404  3.475552  8.211853  3.219150   \n",
       "GSM3188464  8.679320  6.507502  5.760880  3.430780  7.947054  3.171273   \n",
       "GSM3188465  7.967007  6.869358  6.155583  2.740770  8.183173  3.346364   \n",
       "GSM3188466  8.517387  6.733282  5.546450  2.678560  8.090283  3.457542   \n",
       "\n",
       "              ZYG11B     ZZEF1      ZZZ3  \n",
       "GSM3188436  6.195087  7.666162  6.685938  \n",
       "GSM3188437  6.139374  7.575070  6.360326  \n",
       "GSM3188438  6.160805  7.734392  6.078101  \n",
       "GSM3188439  6.137104  7.540814  6.454944  \n",
       "GSM3188440  5.938086  7.502443  6.175240  \n",
       "GSM3188441  5.975827  7.580190  6.206856  \n",
       "GSM3188442  6.044921  7.583205  6.291254  \n",
       "GSM3188443  6.076884  7.574004  6.533650  \n",
       "GSM3188444  6.343047  7.621562  6.387210  \n",
       "GSM3188445  5.914579  7.645101  6.051843  \n",
       "GSM3188446  5.954905  7.621656  5.828861  \n",
       "GSM3188447  6.221065  7.580450  6.350554  \n",
       "GSM3188448  6.239084  7.604069  6.519019  \n",
       "GSM3188449  6.158042  7.580968  6.462511  \n",
       "GSM3188450  6.138881  7.593875  6.394661  \n",
       "GSM3188451  6.227980  7.640666  6.453943  \n",
       "GSM3188452  6.254845  7.654089  6.585905  \n",
       "GSM3188453  6.068423  7.752699  6.145825  \n",
       "GSM3188454  6.021599  7.611209  6.227735  \n",
       "GSM3188455  6.256100  7.665015  6.287613  \n",
       "GSM3188456  6.341254  7.618183  6.515540  \n",
       "GSM3188457  6.187632  7.713456  6.297963  \n",
       "GSM3188458  6.020802  7.935561  6.234126  \n",
       "GSM3188459  5.865893  7.679000  6.382137  \n",
       "GSM3188460  5.863600  7.936546  6.318112  \n",
       "GSM3188461  6.145421  7.642127  6.492053  \n",
       "GSM3188462  5.962614  7.820722  6.316718  \n",
       "GSM3188463  5.813261  7.839113  6.431135  \n",
       "GSM3188464  6.128784  7.663779  6.179315  \n",
       "GSM3188465  6.064534  7.741242  6.435414  \n",
       "GSM3188466  5.960804  7.713401  6.401391  \n",
       "\n",
       "[31 rows x 10058 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 31 samples.\n",
      "For the feature 'Coronary artery disease', the least common label is '0.0' with 7 occurrences. This represents 22.58% of the dataset.\n",
      "The distribution of the feature 'Coronary artery disease' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0.0' with 31 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output2\\Jiayi\\Creutzfeldt-Jakob-Disease\\cohort_info.json\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'merged_data' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[19], line 7\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m      6\u001b[0m     save_cohort_info(cohort, JSON_PATH, is_available)\n\u001b[1;32m----> 7\u001b[0m \u001b[43mmerged_data\u001b[49m\u001b[38;5;241m.\u001b[39mhead()\n\u001b[0;32m      8\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_trait_biased:\n\u001b[0;32m      9\u001b[0m     merged_data\u001b[38;5;241m.\u001b[39mto_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, cohort \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m), index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'merged_data' is not defined"
     ]
    }
   ],
   "source": [
    "is_available = False\n",
    "\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
