{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = 'Fibromyalgia'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE67311']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Fibromyalgia\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Fibromyalgia\\\\GSE67311\\\\GSE67311_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Fibromyalgia\\\\GSE67311\\\\GSE67311_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE67311\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Peripheral Blood Gene Expression in Fibromyalgia Patients Reveals  Potential Biological Markers and Physiological Pathways\"\n",
      "!Series_summary\t\"Fibromyalgia (FM) is a common pain disorder characterized by dysregulation in the processing of pain. Although FM has similarities with other rheumatologic pain disorders, the search for objective markers has not been successful. In the current study we analyzed gene expression in the whole blood of 70 fibromyalgia patients and 70 healthy matched controls. Global molecular profiling revealed an upregulation of several inflammatory molecules in FM patients and downregulation of specific pathways related to hypersensitivity and allergy. There was a differential expression of genes in known pathways for pain processing, such as glutamine/glutamate signaling and axonal development. We also identified a panel of candidate gene expression-based classifiers that could establish an objective blood-based molecular diagnostic to objectively identify FM patients and guide design and testing of new therapies. Ten classifier probesets (CPA3, C11orf83, LOC100131943, RGS17, PARD3B, ANKRD20A9P, TTLL7, C8orf12, KAT2B and RIOK3) provided a diagnostic sensitivity of 95% and a specificity of 96%. Molecular scores developed from these classifiers were able to clearly distinguish FM patients from healthy controls. An understanding of molecular dysregulation in fibromyalgia is in its infancy; however the results described herein indicate blood global gene expression profiling provides many testable hypotheses that deserve further exploration.\"\n",
      "!Series_overall_design\t\"Blood samples were collected in PAXgene tubes and collected samples were stored at -80oC.  The RNA was isolated using the PAXgene RNA isolation kit according to standard protocols. Total RNA was quantified on a Nanodrop spectrophotometer and visualized for quality on an Agilent Bioanalyzer. Samples with an average RIN (RNA Integrity Number) >8, indicating good quality RNA, were processed. 200ng of total RNA was amplified and then hybridized to Affymetrix® Human Gene 1.1 ST Peg arrays using standard manufacturer’s protocols on a Gene Titan MC instrument. Data was analyzed using Partek Genomics Suite (version 6.6) using RMA normalization. All genes with Log2 signal intensity less than 4.8 were excluded from analysis based on low expression. Differential expression analysis was carried out using a one way ANOVA by using Method of Moments and Fisher's Least Significant Difference (LSD) tests with a threshold p-value <0.005 for the biological and molecular function analyses, and a more conservative threshold p-value <0.001 (FDR q-value range 0.002% to 13%) for candidate diagnostic signatures.\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1644447</th>\n",
       "      <th>GSM1644448</th>\n",
       "      <th>GSM1644449</th>\n",
       "      <th>GSM1644450</th>\n",
       "      <th>GSM1644451</th>\n",
       "      <th>GSM1644452</th>\n",
       "      <th>GSM1644453</th>\n",
       "      <th>GSM1644454</th>\n",
       "      <th>GSM1644455</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1644579</th>\n",
       "      <th>GSM1644580</th>\n",
       "      <th>GSM1644581</th>\n",
       "      <th>GSM1644582</th>\n",
       "      <th>GSM1644583</th>\n",
       "      <th>GSM1644584</th>\n",
       "      <th>GSM1644585</th>\n",
       "      <th>GSM1644586</th>\n",
       "      <th>GSM1644587</th>\n",
       "      <th>GSM1644588</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>diagnosis: healthy control</td>\n",
       "      <td>...</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "      <td>diagnosis: fibromyalgia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "      <td>tissue: peripheral blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>fiqr score: 8.5</td>\n",
       "      <td>fiqr score: -2.0</td>\n",
       "      <td>fiqr score: 9.8</td>\n",
       "      <td>fiqr score: 0.5</td>\n",
       "      <td>fiqr score: -1.0</td>\n",
       "      <td>fiqr score: -0.5</td>\n",
       "      <td>fiqr score: 2.2</td>\n",
       "      <td>fiqr score: -2.0</td>\n",
       "      <td>fiqr score: -2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>fiqr score: 41.8</td>\n",
       "      <td>fiqr score: 54.5</td>\n",
       "      <td>fiqr score: 63.0</td>\n",
       "      <td>fiqr score: 64.0</td>\n",
       "      <td>fiqr score: 67.2</td>\n",
       "      <td>fiqr score: 17.8</td>\n",
       "      <td>fiqr score: 17.8</td>\n",
       "      <td>fiqr score: 41.7</td>\n",
       "      <td>fiqr score: 81.2</td>\n",
       "      <td>fiqr score: 54.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>bmi: 36</td>\n",
       "      <td>bmi: 34</td>\n",
       "      <td>bmi: 33</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 24</td>\n",
       "      <td>bmi: 28</td>\n",
       "      <td>bmi: 23</td>\n",
       "      <td>bmi: 48</td>\n",
       "      <td>bmi: 48</td>\n",
       "      <td>...</td>\n",
       "      <td>bmi: 22</td>\n",
       "      <td>bmi: 25</td>\n",
       "      <td>bmi: 29</td>\n",
       "      <td>bmi: 38</td>\n",
       "      <td>bmi: 28</td>\n",
       "      <td>bmi: 37</td>\n",
       "      <td>bmi: 0</td>\n",
       "      <td>bmi: 38</td>\n",
       "      <td>bmi: 20</td>\n",
       "      <td>bmi: 34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>...</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: No</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: -</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: Yes</td>\n",
       "      <td>migraine: No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 143 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                  GSM1644447  \\\n",
       "0  !Sample_characteristics_ch1  diagnosis: healthy control   \n",
       "1  !Sample_characteristics_ch1    tissue: peripheral blood   \n",
       "2  !Sample_characteristics_ch1             fiqr score: 8.5   \n",
       "3  !Sample_characteristics_ch1                     bmi: 36   \n",
       "4  !Sample_characteristics_ch1                migraine: No   \n",
       "\n",
       "                   GSM1644448                  GSM1644449  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood   \n",
       "2            fiqr score: -2.0             fiqr score: 9.8   \n",
       "3                     bmi: 34                     bmi: 33   \n",
       "4                migraine: No                migraine: No   \n",
       "\n",
       "                   GSM1644450                  GSM1644451  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood   \n",
       "2             fiqr score: 0.5            fiqr score: -1.0   \n",
       "3                     bmi: 22                     bmi: 24   \n",
       "4                migraine: No                migraine: No   \n",
       "\n",
       "                   GSM1644452                  GSM1644453  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood   \n",
       "2            fiqr score: -0.5             fiqr score: 2.2   \n",
       "3                     bmi: 28                     bmi: 23   \n",
       "4                migraine: No                migraine: No   \n",
       "\n",
       "                   GSM1644454                  GSM1644455  ...  \\\n",
       "0  diagnosis: healthy control  diagnosis: healthy control  ...   \n",
       "1    tissue: peripheral blood    tissue: peripheral blood  ...   \n",
       "2            fiqr score: -2.0            fiqr score: -2.0  ...   \n",
       "3                     bmi: 48                     bmi: 48  ...   \n",
       "4                migraine: No                migraine: No  ...   \n",
       "\n",
       "                 GSM1644579                GSM1644580  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 41.8          fiqr score: 54.5   \n",
       "3                   bmi: 22                   bmi: 25   \n",
       "4              migraine: No             migraine: Yes   \n",
       "\n",
       "                 GSM1644581                GSM1644582  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 63.0          fiqr score: 64.0   \n",
       "3                   bmi: 29                   bmi: 38   \n",
       "4             migraine: Yes              migraine: No   \n",
       "\n",
       "                 GSM1644583                GSM1644584  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 67.2          fiqr score: 17.8   \n",
       "3                   bmi: 28                   bmi: 37   \n",
       "4             migraine: Yes             migraine: Yes   \n",
       "\n",
       "                 GSM1644585                GSM1644586  \\\n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia   \n",
       "1  tissue: peripheral blood  tissue: peripheral blood   \n",
       "2          fiqr score: 17.8          fiqr score: 41.7   \n",
       "3                    bmi: 0                   bmi: 38   \n",
       "4               migraine: -             migraine: Yes   \n",
       "\n",
       "                 GSM1644587                GSM1644588  \n",
       "0   diagnosis: fibromyalgia   diagnosis: fibromyalgia  \n",
       "1  tissue: peripheral blood  tissue: peripheral blood  \n",
       "2          fiqr score: 81.2          fiqr score: 54.7  \n",
       "3                   bmi: 20                   bmi: 34  \n",
       "4             migraine: Yes              migraine: No  \n",
       "\n",
       "[5 rows x 143 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'diagnosis: healthy control',\n",
       "       'diagnosis: fibromyalgia'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[0]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    # Check if the trait indicates a normal disease status\n",
    "    if trait == 'diagnosis: healthy control':\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(':')[1])/12\n",
    "        age = round(age, 0)\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1644447</th>\n",
       "      <th>GSM1644448</th>\n",
       "      <th>GSM1644449</th>\n",
       "      <th>GSM1644450</th>\n",
       "      <th>GSM1644451</th>\n",
       "      <th>GSM1644452</th>\n",
       "      <th>GSM1644453</th>\n",
       "      <th>GSM1644454</th>\n",
       "      <th>GSM1644455</th>\n",
       "      <th>GSM1644456</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1644579</th>\n",
       "      <th>GSM1644580</th>\n",
       "      <th>GSM1644581</th>\n",
       "      <th>GSM1644582</th>\n",
       "      <th>GSM1644583</th>\n",
       "      <th>GSM1644584</th>\n",
       "      <th>GSM1644585</th>\n",
       "      <th>GSM1644586</th>\n",
       "      <th>GSM1644587</th>\n",
       "      <th>GSM1644588</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Fibromyalgia</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM1644447  GSM1644448  GSM1644449  GSM1644450  GSM1644451  \\\n",
       "Fibromyalgia           0           0           0           0           0   \n",
       "\n",
       "              GSM1644452  GSM1644453  GSM1644454  GSM1644455  GSM1644456  ...  \\\n",
       "Fibromyalgia           0           0           0           0           0  ...   \n",
       "\n",
       "              GSM1644579  GSM1644580  GSM1644581  GSM1644582  GSM1644583  \\\n",
       "Fibromyalgia           1           1           1           1           1   \n",
       "\n",
       "              GSM1644584  GSM1644585  GSM1644586  GSM1644587  GSM1644588  \n",
       "Fibromyalgia           1           1           1           1           1  \n",
       "\n",
       "[1 rows x 142 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1644447</th>\n",
       "      <th>GSM1644448</th>\n",
       "      <th>GSM1644449</th>\n",
       "      <th>GSM1644450</th>\n",
       "      <th>GSM1644451</th>\n",
       "      <th>GSM1644452</th>\n",
       "      <th>GSM1644453</th>\n",
       "      <th>GSM1644454</th>\n",
       "      <th>GSM1644455</th>\n",
       "      <th>GSM1644456</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1644579</th>\n",
       "      <th>GSM1644580</th>\n",
       "      <th>GSM1644581</th>\n",
       "      <th>GSM1644582</th>\n",
       "      <th>GSM1644583</th>\n",
       "      <th>GSM1644584</th>\n",
       "      <th>GSM1644585</th>\n",
       "      <th>GSM1644586</th>\n",
       "      <th>GSM1644587</th>\n",
       "      <th>GSM1644588</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7892501</th>\n",
       "      <td>5.62341</td>\n",
       "      <td>4.54841</td>\n",
       "      <td>4.74053</td>\n",
       "      <td>3.06227</td>\n",
       "      <td>3.65178</td>\n",
       "      <td>4.34336</td>\n",
       "      <td>5.36535</td>\n",
       "      <td>3.69126</td>\n",
       "      <td>2.52748</td>\n",
       "      <td>4.45481</td>\n",
       "      <td>...</td>\n",
       "      <td>2.93373</td>\n",
       "      <td>4.78499</td>\n",
       "      <td>2.86791</td>\n",
       "      <td>5.63311</td>\n",
       "      <td>4.38304</td>\n",
       "      <td>3.72055</td>\n",
       "      <td>3.03068</td>\n",
       "      <td>4.61205</td>\n",
       "      <td>5.38301</td>\n",
       "      <td>3.56539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7892502</th>\n",
       "      <td>5.37542</td>\n",
       "      <td>4.78069</td>\n",
       "      <td>5.70991</td>\n",
       "      <td>5.38193</td>\n",
       "      <td>5.92017</td>\n",
       "      <td>6.09522</td>\n",
       "      <td>5.78167</td>\n",
       "      <td>4.59286</td>\n",
       "      <td>5.16439</td>\n",
       "      <td>5.94071</td>\n",
       "      <td>...</td>\n",
       "      <td>5.68648</td>\n",
       "      <td>5.48387</td>\n",
       "      <td>5.50597</td>\n",
       "      <td>5.27264</td>\n",
       "      <td>6.12783</td>\n",
       "      <td>6.24468</td>\n",
       "      <td>4.73856</td>\n",
       "      <td>5.35777</td>\n",
       "      <td>5.66831</td>\n",
       "      <td>6.07197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7892503</th>\n",
       "      <td>5.14609</td>\n",
       "      <td>4.74459</td>\n",
       "      <td>5.57936</td>\n",
       "      <td>4.72783</td>\n",
       "      <td>4.83541</td>\n",
       "      <td>5.62107</td>\n",
       "      <td>5.26481</td>\n",
       "      <td>4.34715</td>\n",
       "      <td>3.71148</td>\n",
       "      <td>4.25644</td>\n",
       "      <td>...</td>\n",
       "      <td>5.23848</td>\n",
       "      <td>4.28919</td>\n",
       "      <td>4.21250</td>\n",
       "      <td>4.37036</td>\n",
       "      <td>5.32584</td>\n",
       "      <td>4.36270</td>\n",
       "      <td>5.43636</td>\n",
       "      <td>5.27041</td>\n",
       "      <td>5.41011</td>\n",
       "      <td>6.01511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7892504</th>\n",
       "      <td>9.50803</td>\n",
       "      <td>9.64513</td>\n",
       "      <td>9.51809</td>\n",
       "      <td>9.20097</td>\n",
       "      <td>9.20887</td>\n",
       "      <td>9.34666</td>\n",
       "      <td>9.15673</td>\n",
       "      <td>10.11560</td>\n",
       "      <td>9.40959</td>\n",
       "      <td>9.49014</td>\n",
       "      <td>...</td>\n",
       "      <td>9.02612</td>\n",
       "      <td>9.15626</td>\n",
       "      <td>9.21809</td>\n",
       "      <td>8.86528</td>\n",
       "      <td>9.47661</td>\n",
       "      <td>10.31210</td>\n",
       "      <td>9.50119</td>\n",
       "      <td>9.54925</td>\n",
       "      <td>9.48669</td>\n",
       "      <td>8.83406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7892505</th>\n",
       "      <td>3.15360</td>\n",
       "      <td>3.26439</td>\n",
       "      <td>4.35030</td>\n",
       "      <td>2.60544</td>\n",
       "      <td>3.78148</td>\n",
       "      <td>4.04654</td>\n",
       "      <td>5.30473</td>\n",
       "      <td>3.62526</td>\n",
       "      <td>2.82275</td>\n",
       "      <td>2.74823</td>\n",
       "      <td>...</td>\n",
       "      <td>2.64208</td>\n",
       "      <td>2.78818</td>\n",
       "      <td>3.98362</td>\n",
       "      <td>3.54878</td>\n",
       "      <td>2.34695</td>\n",
       "      <td>3.96788</td>\n",
       "      <td>3.31853</td>\n",
       "      <td>2.57307</td>\n",
       "      <td>3.19400</td>\n",
       "      <td>3.01890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 142 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM1644447  GSM1644448  GSM1644449  GSM1644450  GSM1644451  \\\n",
       "ID                                                                    \n",
       "7892501     5.62341     4.54841     4.74053     3.06227     3.65178   \n",
       "7892502     5.37542     4.78069     5.70991     5.38193     5.92017   \n",
       "7892503     5.14609     4.74459     5.57936     4.72783     4.83541   \n",
       "7892504     9.50803     9.64513     9.51809     9.20097     9.20887   \n",
       "7892505     3.15360     3.26439     4.35030     2.60544     3.78148   \n",
       "\n",
       "         GSM1644452  GSM1644453  GSM1644454  GSM1644455  GSM1644456  ...  \\\n",
       "ID                                                                   ...   \n",
       "7892501     4.34336     5.36535     3.69126     2.52748     4.45481  ...   \n",
       "7892502     6.09522     5.78167     4.59286     5.16439     5.94071  ...   \n",
       "7892503     5.62107     5.26481     4.34715     3.71148     4.25644  ...   \n",
       "7892504     9.34666     9.15673    10.11560     9.40959     9.49014  ...   \n",
       "7892505     4.04654     5.30473     3.62526     2.82275     2.74823  ...   \n",
       "\n",
       "         GSM1644579  GSM1644580  GSM1644581  GSM1644582  GSM1644583  \\\n",
       "ID                                                                    \n",
       "7892501     2.93373     4.78499     2.86791     5.63311     4.38304   \n",
       "7892502     5.68648     5.48387     5.50597     5.27264     6.12783   \n",
       "7892503     5.23848     4.28919     4.21250     4.37036     5.32584   \n",
       "7892504     9.02612     9.15626     9.21809     8.86528     9.47661   \n",
       "7892505     2.64208     2.78818     3.98362     3.54878     2.34695   \n",
       "\n",
       "         GSM1644584  GSM1644585  GSM1644586  GSM1644587  GSM1644588  \n",
       "ID                                                                   \n",
       "7892501     3.72055     3.03068     4.61205     5.38301     3.56539  \n",
       "7892502     6.24468     4.73856     5.35777     5.66831     6.07197  \n",
       "7892503     4.36270     5.43636     5.27041     5.41011     6.01511  \n",
       "7892504    10.31210     9.50119     9.54925     9.48669     8.83406  \n",
       "7892505     3.96788     3.31853     2.57307     3.19400     3.01890  \n",
       "\n",
       "[5 rows x 142 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['7896736', '7896738', '7896740', '7896742', '7896744'], 'GB_LIST': [nan, nan, 'NM_001005240,NM_001004195,NM_001005484,BC136848,BC136907', 'BC118988,AL137655', 'NM_001005277,NM_001005221,NM_001005224,NM_001005504,BC137547'], 'SPOT_ID': ['chr1:53049-54936', 'chr1:63015-63887', 'chr1:69091-70008', 'chr1:334129-334296', 'chr1:367659-368597'], 'seqname': ['chr1', 'chr1', 'chr1', 'chr1', 'chr1'], 'RANGE_GB': ['NC_000001.10', 'NC_000001.10', 'NC_000001.10', 'NC_000001.10', 'NC_000001.10'], 'RANGE_STRAND': ['+', '+', '+', '+', '+'], 'RANGE_START': [53049.0, 63015.0, 69091.0, 334129.0, 367659.0], 'RANGE_STOP': [54936.0, 63887.0, 70008.0, 334296.0, 368597.0], 'total_probes': [7.0, 31.0, 24.0, 6.0, 36.0], 'gene_assignment': ['---', '---', 'NM_001005240 // OR4F17 // olfactory receptor, family 4, subfamily F, member 17 // 19p13.3 // 81099 /// NM_001004195 // OR4F4 // olfactory receptor, family 4, subfamily F, member 4 // 15q26.3 // 26682 /// NM_001005484 // OR4F5 // olfactory receptor, family 4, subfamily F, member 5 // 1p36.33 // 79501 /// ENST00000318050 // OR4F17 // olfactory receptor, family 4, subfamily F, member 17 // 19p13.3 // 81099 /// ENST00000335137 // OR4F4 // olfactory receptor, family 4, subfamily F, member 4 // 15q26.3 // 26682 /// ENST00000326183 // OR4F5 // olfactory receptor, family 4, subfamily F, member 5 // 1p36.33 // 79501 /// BC136848 // OR4F17 // olfactory receptor, family 4, subfamily F, member 17 // 19p13.3 // 81099 /// BC136907 // OR4F4 // olfactory receptor, family 4, subfamily F, member 4 // 15q26.3 // 26682 /// ENST00000442916 // OR4F17 // olfactory receptor, family 4, subfamily F, member 17 // 19p13.3 // 81099', 'ENST00000388975 // SEPT14 // septin 14 // 7p11.2 // 346288 /// BC118988 // NCRNA00266 // non-protein coding RNA 266 // --- // 140849 /// AL137655 // LOC100134822 // similar to hCG1739109 // --- // 100134822', 'NM_001005277 // OR4F16 // olfactory receptor, family 4, subfamily F, member 16 // 1p36.33 // 81399 /// NM_001005221 // OR4F29 // olfactory receptor, family 4, subfamily F, member 29 // 1p36.33 // 729759 /// NM_001005224 // OR4F3 // olfactory receptor, family 4, subfamily F, member 3 // 5q35.3 // 26683 /// NM_001005504 // OR4F21 // olfactory receptor, family 4, subfamily F, member 21 // 8p23.3 // 441308 /// ENST00000320901 // OR4F21 // olfactory receptor, family 4, subfamily F, member 21 // 8p23.3 // 441308 /// BC137547 // OR4F3 // olfactory receptor, family 4, subfamily F, member 3 // 5q35.3 // 26683 /// BC137547 // OR4F16 // olfactory receptor, family 4, subfamily F, member 16 // 1p36.33 // 81399 /// BC137547 // OR4F29 // olfactory receptor, family 4, subfamily F, member 29 // 1p36.33 // 729759'], 'mrna_assignment': ['---', 'ENST00000328113 // ENSEMBL // cdna:pseudogene chromosome:GRCh37:15:102467008:102467910:-1 gene:ENSG00000183909 // chr1 // 100 // 100 // 31 // 31 // 0 /// ENST00000318181 // ENSEMBL // cdna:pseudogene chromosome:GRCh37:19:104601:105256:1 gene:ENSG00000176705 // chr1 // 100 // 100 // 31 // 31 // 0 /// ENST00000492842 // ENSEMBL // cdna:pseudogene chromosome:GRCh37:1:62948:63887:1 gene:ENSG00000240361 // chr1 // 100 // 100 // 31 // 31 // 0', 'NM_001005240 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 17 (OR4F17), mRNA. // chr1 // 100 // 100 // 24 // 24 // 0 /// NM_001004195 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 4 (OR4F4), mRNA. // chr1 // 100 // 100 // 24 // 24 // 0 /// NM_001005484 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 5 (OR4F5), mRNA. // chr1 // 100 // 100 // 24 // 24 // 0 /// ENST00000318050 // ENSEMBL // Olfactory receptor 4F17 gene:ENSG00000176695 // chr1 // 100 // 100 // 24 // 24 // 0 /// ENST00000335137 // ENSEMBL // Olfactory receptor 4F4 gene:ENSG00000186092 // chr1 // 100 // 100 // 24 // 24 // 0 /// ENST00000326183 // ENSEMBL // Olfactory receptor 4F5 gene:ENSG00000177693 // chr1 // 100 // 100 // 24 // 24 // 0 /// BC136848 // GenBank // Homo sapiens olfactory receptor, family 4, subfamily F, member 17, mRNA (cDNA clone MGC:168462 IMAGE:9020839), complete cds. // chr1 // 100 // 100 // 24 // 24 // 0 /// BC136907 // GenBank // Homo sapiens olfactory receptor, family 4, subfamily F, member 4, mRNA (cDNA clone MGC:168521 IMAGE:9020898), complete cds. // chr1 // 100 // 100 // 24 // 24 // 0 /// ENST00000442916 // ENSEMBL // OR4F4 (Fragment) gene:ENSG00000176695 // chr1 // 100 // 88 // 21 // 21 // 0', 'ENST00000388975 // ENSEMBL // Septin-14 gene:ENSG00000154997 // chr1 // 50 // 100 // 3 // 6 // 0 /// BC118988 // GenBank // Homo sapiens chromosome 20 open reading frame 69, mRNA (cDNA clone MGC:141807 IMAGE:40035995), complete cds. // chr1 // 100 // 100 // 6 // 6 // 0 /// AL137655 // GenBank // Homo sapiens mRNA; cDNA DKFZp434B2016 (from clone DKFZp434B2016). // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000428915 // ENSEMBL // cdna:known chromosome:GRCh37:10:38742109:38755311:1 gene:ENSG00000203496 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000455207 // ENSEMBL // cdna:known chromosome:GRCh37:1:334129:446155:1 gene:ENSG00000224813 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000455464 // ENSEMBL // cdna:known chromosome:GRCh37:1:334140:342806:1 gene:ENSG00000224813 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000440200 // ENSEMBL // cdna:known chromosome:GRCh37:1:536816:655580:-1 gene:ENSG00000230021 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000279067 // ENSEMBL // cdna:known chromosome:GRCh37:20:62921738:62934912:1 gene:ENSG00000149656 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000499986 // ENSEMBL // cdna:known chromosome:GRCh37:5:180717576:180761371:1 gene:ENSG00000248628 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000436899 // ENSEMBL // cdna:known chromosome:GRCh37:6:131910:144885:-1 gene:ENSG00000170590 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000432557 // ENSEMBL // cdna:known chromosome:GRCh37:8:132324:150572:-1 gene:ENSG00000250210 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000523795 // ENSEMBL // cdna:known chromosome:GRCh37:8:141690:150563:-1 gene:ENSG00000250210 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000490482 // ENSEMBL // cdna:known chromosome:GRCh37:8:149942:163324:-1 gene:ENSG00000223508 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000307499 // ENSEMBL // cdna:known supercontig::GL000227.1:57780:70752:-1 gene:ENSG00000229450 // chr1 // 100 // 100 // 6 // 6 // 0 /// ENST00000441245 // ENSEMBL // cdna:known chromosome:GRCh37:1:637316:655530:-1 gene:ENSG00000230021 // chr1 // 100 // 67 // 4 // 4 // 0 /// ENST00000425473 // ENSEMBL // cdna:known chromosome:GRCh37:20:62926294:62944485:1 gene:ENSG00000149656 // chr1 // 100 // 67 // 4 // 4 // 0 /// ENST00000471248 // ENSEMBL // cdna:known chromosome:GRCh37:1:110953:129173:-1 gene:ENSG00000238009 // chr1 // 75 // 67 // 3 // 4 // 0', 'NM_001005277 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 16 (OR4F16), mRNA. // chr1 // 100 // 100 // 36 // 36 // 0 /// NM_001005221 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 29 (OR4F29), mRNA. // chr1 // 100 // 100 // 36 // 36 // 0 /// NM_001005224 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 3 (OR4F3), mRNA. // chr1 // 100 // 100 // 36 // 36 // 0 /// NM_001005504 // RefSeq // Homo sapiens olfactory receptor, family 4, subfamily F, member 21 (OR4F21), mRNA. // chr1 // 89 // 100 // 32 // 36 // 0 /// ENST00000320901 // ENSEMBL // Olfactory receptor 4F21 gene:ENSG00000176269 // chr1 // 89 // 100 // 32 // 36 // 0 /// BC137547 // GenBank // Homo sapiens olfactory receptor, family 4, subfamily F, member 3, mRNA (cDNA clone MGC:169170 IMAGE:9021547), complete cds. // chr1 // 100 // 100 // 36 // 36 // 0 /// ENST00000426406 // ENSEMBL // cdna:known chromosome:GRCh37:1:367640:368634:1 gene:ENSG00000235249 // chr1 // 100 // 100 // 36 // 36 // 0 /// ENST00000332831 // ENSEMBL // cdna:known chromosome:GRCh37:1:621096:622034:-1 gene:ENSG00000185097 // chr1 // 100 // 100 // 36 // 36 // 0 /// ENST00000456475 // ENSEMBL // cdna:known chromosome:GRCh37:5:180794269:180795263:1 gene:ENSG00000230178 // chr1 // 100 // 100 // 36 // 36 // 0 /// ENST00000521196 // ENSEMBL // cdna:known chromosome:GRCh37:11:86612:87605:-1 gene:ENSG00000224777 // chr1 // 78 // 100 // 28 // 36 // 0'], 'category': ['---', 'main', 'main', 'main', 'main']}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_LIST', 'SPOT_ID', 'seqname', 'RANGE_GB', 'RANGE_STRAND',\n",
       "       'RANGE_START', 'RANGE_STOP', 'total_probes', 'gene_assignment',\n",
       "       'mrna_assignment', 'category'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1054461</th>\n",
       "      <th>GSM1054462</th>\n",
       "      <th>GSM1054463</th>\n",
       "      <th>GSM1054464</th>\n",
       "      <th>GSM1054465</th>\n",
       "      <th>GSM1054466</th>\n",
       "      <th>GSM1054467</th>\n",
       "      <th>GSM1054468</th>\n",
       "      <th>GSM1054469</th>\n",
       "      <th>GSM1054470</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1054504</th>\n",
       "      <th>GSM1054505</th>\n",
       "      <th>GSM1054506</th>\n",
       "      <th>GSM1054507</th>\n",
       "      <th>GSM1054508</th>\n",
       "      <th>GSM1054509</th>\n",
       "      <th>GSM1054510</th>\n",
       "      <th>GSM1054511</th>\n",
       "      <th>GSM1054512</th>\n",
       "      <th>GSM1054513</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>6.1013</td>\n",
       "      <td>5.8975</td>\n",
       "      <td>6.1725</td>\n",
       "      <td>5.9452</td>\n",
       "      <td>5.8125</td>\n",
       "      <td>6.2270</td>\n",
       "      <td>6.6265</td>\n",
       "      <td>6.1640</td>\n",
       "      <td>6.2533</td>\n",
       "      <td>5.9966</td>\n",
       "      <td>...</td>\n",
       "      <td>8.1697</td>\n",
       "      <td>7.3470</td>\n",
       "      <td>7.3858</td>\n",
       "      <td>7.2936</td>\n",
       "      <td>7.3586</td>\n",
       "      <td>7.5003</td>\n",
       "      <td>7.7289</td>\n",
       "      <td>7.2938</td>\n",
       "      <td>7.2880</td>\n",
       "      <td>7.3009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>4.1155</td>\n",
       "      <td>3.8231</td>\n",
       "      <td>4.4809</td>\n",
       "      <td>3.8832</td>\n",
       "      <td>3.5292</td>\n",
       "      <td>4.0760</td>\n",
       "      <td>4.2817</td>\n",
       "      <td>4.0797</td>\n",
       "      <td>3.7546</td>\n",
       "      <td>3.7637</td>\n",
       "      <td>...</td>\n",
       "      <td>6.9534</td>\n",
       "      <td>6.8703</td>\n",
       "      <td>6.5639</td>\n",
       "      <td>7.2063</td>\n",
       "      <td>6.9356</td>\n",
       "      <td>6.7637</td>\n",
       "      <td>7.2619</td>\n",
       "      <td>6.9173</td>\n",
       "      <td>6.9877</td>\n",
       "      <td>6.8485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>3.0490</td>\n",
       "      <td>2.8329</td>\n",
       "      <td>3.0462</td>\n",
       "      <td>3.0998</td>\n",
       "      <td>2.8448</td>\n",
       "      <td>2.9790</td>\n",
       "      <td>3.1286</td>\n",
       "      <td>3.2239</td>\n",
       "      <td>2.9757</td>\n",
       "      <td>2.8277</td>\n",
       "      <td>...</td>\n",
       "      <td>3.8526</td>\n",
       "      <td>3.1797</td>\n",
       "      <td>3.2482</td>\n",
       "      <td>3.1539</td>\n",
       "      <td>3.3267</td>\n",
       "      <td>3.2850</td>\n",
       "      <td>3.4254</td>\n",
       "      <td>3.2007</td>\n",
       "      <td>3.3306</td>\n",
       "      <td>3.1708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>9.5803</td>\n",
       "      <td>8.8574</td>\n",
       "      <td>8.6237</td>\n",
       "      <td>8.7689</td>\n",
       "      <td>9.6166</td>\n",
       "      <td>9.5560</td>\n",
       "      <td>9.5490</td>\n",
       "      <td>8.8824</td>\n",
       "      <td>8.9636</td>\n",
       "      <td>8.7677</td>\n",
       "      <td>...</td>\n",
       "      <td>8.3536</td>\n",
       "      <td>8.4455</td>\n",
       "      <td>6.4056</td>\n",
       "      <td>5.0722</td>\n",
       "      <td>7.2323</td>\n",
       "      <td>6.9800</td>\n",
       "      <td>8.2093</td>\n",
       "      <td>5.0774</td>\n",
       "      <td>5.0126</td>\n",
       "      <td>10.3406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>3.2303</td>\n",
       "      <td>3.1281</td>\n",
       "      <td>3.3229</td>\n",
       "      <td>3.3928</td>\n",
       "      <td>3.1351</td>\n",
       "      <td>3.3229</td>\n",
       "      <td>3.7033</td>\n",
       "      <td>3.3818</td>\n",
       "      <td>3.2558</td>\n",
       "      <td>3.2793</td>\n",
       "      <td>...</td>\n",
       "      <td>4.7692</td>\n",
       "      <td>3.7672</td>\n",
       "      <td>3.9695</td>\n",
       "      <td>3.8859</td>\n",
       "      <td>3.8911</td>\n",
       "      <td>4.0563</td>\n",
       "      <td>4.1931</td>\n",
       "      <td>3.9923</td>\n",
       "      <td>3.7533</td>\n",
       "      <td>4.0087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>3.2158</td>\n",
       "      <td>3.7873</td>\n",
       "      <td>3.4058</td>\n",
       "      <td>3.3421</td>\n",
       "      <td>3.3949</td>\n",
       "      <td>3.2535</td>\n",
       "      <td>3.6718</td>\n",
       "      <td>3.5500</td>\n",
       "      <td>3.4669</td>\n",
       "      <td>3.2070</td>\n",
       "      <td>...</td>\n",
       "      <td>4.8983</td>\n",
       "      <td>4.7218</td>\n",
       "      <td>4.8042</td>\n",
       "      <td>4.0250</td>\n",
       "      <td>4.4530</td>\n",
       "      <td>4.5717</td>\n",
       "      <td>4.6689</td>\n",
       "      <td>4.7296</td>\n",
       "      <td>4.4920</td>\n",
       "      <td>4.3177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>9.0668</td>\n",
       "      <td>9.4642</td>\n",
       "      <td>9.3095</td>\n",
       "      <td>9.4767</td>\n",
       "      <td>8.9353</td>\n",
       "      <td>8.6806</td>\n",
       "      <td>8.8417</td>\n",
       "      <td>8.3534</td>\n",
       "      <td>8.3751</td>\n",
       "      <td>8.3136</td>\n",
       "      <td>...</td>\n",
       "      <td>9.4790</td>\n",
       "      <td>9.6251</td>\n",
       "      <td>9.4461</td>\n",
       "      <td>9.4394</td>\n",
       "      <td>9.2585</td>\n",
       "      <td>9.0012</td>\n",
       "      <td>9.2760</td>\n",
       "      <td>9.3080</td>\n",
       "      <td>9.1942</td>\n",
       "      <td>9.4302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>7.1701</td>\n",
       "      <td>6.9826</td>\n",
       "      <td>6.9313</td>\n",
       "      <td>7.3664</td>\n",
       "      <td>7.2823</td>\n",
       "      <td>6.6743</td>\n",
       "      <td>6.9897</td>\n",
       "      <td>7.2487</td>\n",
       "      <td>7.2203</td>\n",
       "      <td>8.2768</td>\n",
       "      <td>...</td>\n",
       "      <td>10.7341</td>\n",
       "      <td>10.9998</td>\n",
       "      <td>10.7922</td>\n",
       "      <td>10.5329</td>\n",
       "      <td>10.4911</td>\n",
       "      <td>10.2403</td>\n",
       "      <td>11.2125</td>\n",
       "      <td>10.5929</td>\n",
       "      <td>10.4838</td>\n",
       "      <td>10.8218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>6.3858</td>\n",
       "      <td>6.5421</td>\n",
       "      <td>6.7484</td>\n",
       "      <td>6.2827</td>\n",
       "      <td>6.6528</td>\n",
       "      <td>6.2846</td>\n",
       "      <td>6.1841</td>\n",
       "      <td>6.1643</td>\n",
       "      <td>6.1597</td>\n",
       "      <td>6.7696</td>\n",
       "      <td>...</td>\n",
       "      <td>7.7777</td>\n",
       "      <td>8.1910</td>\n",
       "      <td>8.3599</td>\n",
       "      <td>8.4375</td>\n",
       "      <td>8.4334</td>\n",
       "      <td>8.3261</td>\n",
       "      <td>7.9002</td>\n",
       "      <td>8.3949</td>\n",
       "      <td>8.2926</td>\n",
       "      <td>8.2379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>5.2251</td>\n",
       "      <td>5.3963</td>\n",
       "      <td>4.8275</td>\n",
       "      <td>5.0468</td>\n",
       "      <td>5.0241</td>\n",
       "      <td>5.1010</td>\n",
       "      <td>4.8599</td>\n",
       "      <td>4.9404</td>\n",
       "      <td>5.0926</td>\n",
       "      <td>5.2231</td>\n",
       "      <td>...</td>\n",
       "      <td>7.3765</td>\n",
       "      <td>7.9798</td>\n",
       "      <td>7.7927</td>\n",
       "      <td>7.9860</td>\n",
       "      <td>7.9013</td>\n",
       "      <td>7.7559</td>\n",
       "      <td>7.3924</td>\n",
       "      <td>8.0305</td>\n",
       "      <td>7.8095</td>\n",
       "      <td>7.6898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17544 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM1054461  GSM1054462  GSM1054463  GSM1054464  GSM1054465  \\\n",
       "A1BG          6.1013      5.8975      6.1725      5.9452      5.8125   \n",
       "A1BG-AS1      4.1155      3.8231      4.4809      3.8832      3.5292   \n",
       "A1CF          3.0490      2.8329      3.0462      3.0998      2.8448   \n",
       "A2M           9.5803      8.8574      8.6237      8.7689      9.6166   \n",
       "A2ML1         3.2303      3.1281      3.3229      3.3928      3.1351   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A        3.2158      3.7873      3.4058      3.3421      3.3949   \n",
       "ZYG11B        9.0668      9.4642      9.3095      9.4767      8.9353   \n",
       "ZYX           7.1701      6.9826      6.9313      7.3664      7.2823   \n",
       "ZZEF1         6.3858      6.5421      6.7484      6.2827      6.6528   \n",
       "ZZZ3          5.2251      5.3963      4.8275      5.0468      5.0241   \n",
       "\n",
       "          GSM1054466  GSM1054467  GSM1054468  GSM1054469  GSM1054470  ...  \\\n",
       "A1BG          6.2270      6.6265      6.1640      6.2533      5.9966  ...   \n",
       "A1BG-AS1      4.0760      4.2817      4.0797      3.7546      3.7637  ...   \n",
       "A1CF          2.9790      3.1286      3.2239      2.9757      2.8277  ...   \n",
       "A2M           9.5560      9.5490      8.8824      8.9636      8.7677  ...   \n",
       "A2ML1         3.3229      3.7033      3.3818      3.2558      3.2793  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A        3.2535      3.6718      3.5500      3.4669      3.2070  ...   \n",
       "ZYG11B        8.6806      8.8417      8.3534      8.3751      8.3136  ...   \n",
       "ZYX           6.6743      6.9897      7.2487      7.2203      8.2768  ...   \n",
       "ZZEF1         6.2846      6.1841      6.1643      6.1597      6.7696  ...   \n",
       "ZZZ3          5.1010      4.8599      4.9404      5.0926      5.2231  ...   \n",
       "\n",
       "          GSM1054504  GSM1054505  GSM1054506  GSM1054507  GSM1054508  \\\n",
       "A1BG          8.1697      7.3470      7.3858      7.2936      7.3586   \n",
       "A1BG-AS1      6.9534      6.8703      6.5639      7.2063      6.9356   \n",
       "A1CF          3.8526      3.1797      3.2482      3.1539      3.3267   \n",
       "A2M           8.3536      8.4455      6.4056      5.0722      7.2323   \n",
       "A2ML1         4.7692      3.7672      3.9695      3.8859      3.8911   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A        4.8983      4.7218      4.8042      4.0250      4.4530   \n",
       "ZYG11B        9.4790      9.6251      9.4461      9.4394      9.2585   \n",
       "ZYX          10.7341     10.9998     10.7922     10.5329     10.4911   \n",
       "ZZEF1         7.7777      8.1910      8.3599      8.4375      8.4334   \n",
       "ZZZ3          7.3765      7.9798      7.7927      7.9860      7.9013   \n",
       "\n",
       "          GSM1054509  GSM1054510  GSM1054511  GSM1054512  GSM1054513  \n",
       "A1BG          7.5003      7.7289      7.2938      7.2880      7.3009  \n",
       "A1BG-AS1      6.7637      7.2619      6.9173      6.9877      6.8485  \n",
       "A1CF          3.2850      3.4254      3.2007      3.3306      3.1708  \n",
       "A2M           6.9800      8.2093      5.0774      5.0126     10.3406  \n",
       "A2ML1         4.0563      4.1931      3.9923      3.7533      4.0087  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "ZYG11A        4.5717      4.6689      4.7296      4.4920      4.3177  \n",
       "ZYG11B        9.0012      9.2760      9.3080      9.1942      9.4302  \n",
       "ZYX          10.2403     11.2125     10.5929     10.4838     10.8218  \n",
       "ZZEF1         8.3261      7.9002      8.3949      8.2926      8.2379  \n",
       "ZZZ3          7.7559      7.3924      8.0305      7.8095      7.6898  \n",
       "\n",
       "[17544 rows x 53 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mitochondrial Disorders</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AA06</th>\n",
       "      <th>AAA1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1054461</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.1013</td>\n",
       "      <td>4.1155</td>\n",
       "      <td>3.0490</td>\n",
       "      <td>9.5803</td>\n",
       "      <td>3.2303</td>\n",
       "      <td>5.7970</td>\n",
       "      <td>3.4187</td>\n",
       "      <td>3.6406</td>\n",
       "      <td>3.0490</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6210</td>\n",
       "      <td>4.6627</td>\n",
       "      <td>6.7564</td>\n",
       "      <td>6.9139</td>\n",
       "      <td>6.4468</td>\n",
       "      <td>3.2158</td>\n",
       "      <td>9.0668</td>\n",
       "      <td>7.1701</td>\n",
       "      <td>6.3858</td>\n",
       "      <td>5.2251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054462</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8975</td>\n",
       "      <td>3.8231</td>\n",
       "      <td>2.8329</td>\n",
       "      <td>8.8574</td>\n",
       "      <td>3.1281</td>\n",
       "      <td>5.2234</td>\n",
       "      <td>4.0981</td>\n",
       "      <td>2.8983</td>\n",
       "      <td>3.5197</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6732</td>\n",
       "      <td>4.0082</td>\n",
       "      <td>6.3447</td>\n",
       "      <td>7.2577</td>\n",
       "      <td>6.3363</td>\n",
       "      <td>3.7873</td>\n",
       "      <td>9.4642</td>\n",
       "      <td>6.9826</td>\n",
       "      <td>6.5421</td>\n",
       "      <td>5.3963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054463</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.1725</td>\n",
       "      <td>4.4809</td>\n",
       "      <td>3.0462</td>\n",
       "      <td>8.6237</td>\n",
       "      <td>3.3229</td>\n",
       "      <td>5.5002</td>\n",
       "      <td>3.6228</td>\n",
       "      <td>4.0165</td>\n",
       "      <td>3.1844</td>\n",
       "      <td>...</td>\n",
       "      <td>3.7046</td>\n",
       "      <td>4.8828</td>\n",
       "      <td>6.4101</td>\n",
       "      <td>6.6484</td>\n",
       "      <td>6.4234</td>\n",
       "      <td>3.4058</td>\n",
       "      <td>9.3095</td>\n",
       "      <td>6.9313</td>\n",
       "      <td>6.7484</td>\n",
       "      <td>4.8275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054464</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.9452</td>\n",
       "      <td>3.8832</td>\n",
       "      <td>3.0998</td>\n",
       "      <td>8.7689</td>\n",
       "      <td>3.3928</td>\n",
       "      <td>5.6048</td>\n",
       "      <td>3.4490</td>\n",
       "      <td>3.0753</td>\n",
       "      <td>3.0915</td>\n",
       "      <td>...</td>\n",
       "      <td>3.3512</td>\n",
       "      <td>4.4825</td>\n",
       "      <td>6.4230</td>\n",
       "      <td>7.1715</td>\n",
       "      <td>6.4230</td>\n",
       "      <td>3.3421</td>\n",
       "      <td>9.4767</td>\n",
       "      <td>7.3664</td>\n",
       "      <td>6.2827</td>\n",
       "      <td>5.0468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054465</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8125</td>\n",
       "      <td>3.5292</td>\n",
       "      <td>2.8448</td>\n",
       "      <td>9.6166</td>\n",
       "      <td>3.1351</td>\n",
       "      <td>5.6190</td>\n",
       "      <td>3.5056</td>\n",
       "      <td>3.2483</td>\n",
       "      <td>2.7159</td>\n",
       "      <td>...</td>\n",
       "      <td>3.8325</td>\n",
       "      <td>4.7506</td>\n",
       "      <td>7.4742</td>\n",
       "      <td>7.4161</td>\n",
       "      <td>6.2950</td>\n",
       "      <td>3.3949</td>\n",
       "      <td>8.9353</td>\n",
       "      <td>7.2823</td>\n",
       "      <td>6.6528</td>\n",
       "      <td>5.0241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054466</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.2270</td>\n",
       "      <td>4.0760</td>\n",
       "      <td>2.9790</td>\n",
       "      <td>9.5560</td>\n",
       "      <td>3.3229</td>\n",
       "      <td>5.6816</td>\n",
       "      <td>3.4400</td>\n",
       "      <td>3.3479</td>\n",
       "      <td>2.8350</td>\n",
       "      <td>...</td>\n",
       "      <td>3.8073</td>\n",
       "      <td>4.5413</td>\n",
       "      <td>6.7705</td>\n",
       "      <td>7.3081</td>\n",
       "      <td>6.4915</td>\n",
       "      <td>3.2535</td>\n",
       "      <td>8.6806</td>\n",
       "      <td>6.6743</td>\n",
       "      <td>6.2846</td>\n",
       "      <td>5.1010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054467</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.6265</td>\n",
       "      <td>4.2817</td>\n",
       "      <td>3.1286</td>\n",
       "      <td>9.5490</td>\n",
       "      <td>3.7033</td>\n",
       "      <td>5.7616</td>\n",
       "      <td>3.8976</td>\n",
       "      <td>3.4092</td>\n",
       "      <td>3.0747</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5066</td>\n",
       "      <td>4.8491</td>\n",
       "      <td>6.7513</td>\n",
       "      <td>7.0802</td>\n",
       "      <td>6.5757</td>\n",
       "      <td>3.6718</td>\n",
       "      <td>8.8417</td>\n",
       "      <td>6.9897</td>\n",
       "      <td>6.1841</td>\n",
       "      <td>4.8599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054468</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.1640</td>\n",
       "      <td>4.0797</td>\n",
       "      <td>3.2239</td>\n",
       "      <td>8.8824</td>\n",
       "      <td>3.3818</td>\n",
       "      <td>5.6999</td>\n",
       "      <td>3.7429</td>\n",
       "      <td>3.4481</td>\n",
       "      <td>3.0651</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5940</td>\n",
       "      <td>4.7115</td>\n",
       "      <td>6.8967</td>\n",
       "      <td>7.3867</td>\n",
       "      <td>6.7240</td>\n",
       "      <td>3.5500</td>\n",
       "      <td>8.3534</td>\n",
       "      <td>7.2487</td>\n",
       "      <td>6.1643</td>\n",
       "      <td>4.9404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054469</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.2533</td>\n",
       "      <td>3.7546</td>\n",
       "      <td>2.9757</td>\n",
       "      <td>8.9636</td>\n",
       "      <td>3.2558</td>\n",
       "      <td>5.5583</td>\n",
       "      <td>4.3190</td>\n",
       "      <td>3.5288</td>\n",
       "      <td>3.3684</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5712</td>\n",
       "      <td>4.4865</td>\n",
       "      <td>6.9326</td>\n",
       "      <td>7.1493</td>\n",
       "      <td>6.7174</td>\n",
       "      <td>3.4669</td>\n",
       "      <td>8.3751</td>\n",
       "      <td>7.2203</td>\n",
       "      <td>6.1597</td>\n",
       "      <td>5.0926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054470</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.9966</td>\n",
       "      <td>3.7637</td>\n",
       "      <td>2.8277</td>\n",
       "      <td>8.7677</td>\n",
       "      <td>3.2793</td>\n",
       "      <td>6.5516</td>\n",
       "      <td>3.4250</td>\n",
       "      <td>3.2469</td>\n",
       "      <td>2.7997</td>\n",
       "      <td>...</td>\n",
       "      <td>3.3897</td>\n",
       "      <td>4.4323</td>\n",
       "      <td>5.6660</td>\n",
       "      <td>6.7189</td>\n",
       "      <td>7.6097</td>\n",
       "      <td>3.2070</td>\n",
       "      <td>8.3136</td>\n",
       "      <td>8.2768</td>\n",
       "      <td>6.7696</td>\n",
       "      <td>5.2231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054471</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.6345</td>\n",
       "      <td>3.6525</td>\n",
       "      <td>2.8254</td>\n",
       "      <td>9.5466</td>\n",
       "      <td>2.9381</td>\n",
       "      <td>6.0614</td>\n",
       "      <td>3.2169</td>\n",
       "      <td>2.9694</td>\n",
       "      <td>2.8535</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6847</td>\n",
       "      <td>4.4523</td>\n",
       "      <td>5.8703</td>\n",
       "      <td>6.9296</td>\n",
       "      <td>7.5199</td>\n",
       "      <td>2.9847</td>\n",
       "      <td>8.2782</td>\n",
       "      <td>8.4515</td>\n",
       "      <td>7.5613</td>\n",
       "      <td>5.2294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054472</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.9679</td>\n",
       "      <td>3.8574</td>\n",
       "      <td>3.0826</td>\n",
       "      <td>9.0008</td>\n",
       "      <td>3.3229</td>\n",
       "      <td>5.7816</td>\n",
       "      <td>4.2293</td>\n",
       "      <td>3.2270</td>\n",
       "      <td>3.5524</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5150</td>\n",
       "      <td>4.5171</td>\n",
       "      <td>6.7663</td>\n",
       "      <td>7.2602</td>\n",
       "      <td>6.5348</td>\n",
       "      <td>3.7670</td>\n",
       "      <td>9.2247</td>\n",
       "      <td>7.3436</td>\n",
       "      <td>6.4235</td>\n",
       "      <td>4.9750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054473</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.6092</td>\n",
       "      <td>3.4124</td>\n",
       "      <td>2.7425</td>\n",
       "      <td>9.7295</td>\n",
       "      <td>2.9885</td>\n",
       "      <td>5.3000</td>\n",
       "      <td>3.5026</td>\n",
       "      <td>2.5168</td>\n",
       "      <td>3.0143</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5778</td>\n",
       "      <td>4.2257</td>\n",
       "      <td>6.6533</td>\n",
       "      <td>6.9227</td>\n",
       "      <td>6.7628</td>\n",
       "      <td>3.2882</td>\n",
       "      <td>10.1225</td>\n",
       "      <td>7.1470</td>\n",
       "      <td>6.8428</td>\n",
       "      <td>5.4319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054474</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.9416</td>\n",
       "      <td>3.9226</td>\n",
       "      <td>2.8950</td>\n",
       "      <td>9.4086</td>\n",
       "      <td>3.1365</td>\n",
       "      <td>5.3408</td>\n",
       "      <td>3.4832</td>\n",
       "      <td>3.1897</td>\n",
       "      <td>2.9552</td>\n",
       "      <td>...</td>\n",
       "      <td>3.7758</td>\n",
       "      <td>4.3914</td>\n",
       "      <td>6.3239</td>\n",
       "      <td>7.0742</td>\n",
       "      <td>6.7333</td>\n",
       "      <td>3.4356</td>\n",
       "      <td>9.6568</td>\n",
       "      <td>7.2235</td>\n",
       "      <td>7.2606</td>\n",
       "      <td>4.7891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054475</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8829</td>\n",
       "      <td>3.5686</td>\n",
       "      <td>2.9108</td>\n",
       "      <td>8.9915</td>\n",
       "      <td>3.2265</td>\n",
       "      <td>5.2228</td>\n",
       "      <td>3.4857</td>\n",
       "      <td>3.9896</td>\n",
       "      <td>3.0070</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5944</td>\n",
       "      <td>4.5104</td>\n",
       "      <td>6.2869</td>\n",
       "      <td>6.9581</td>\n",
       "      <td>6.8440</td>\n",
       "      <td>3.2626</td>\n",
       "      <td>10.0019</td>\n",
       "      <td>7.1317</td>\n",
       "      <td>7.2185</td>\n",
       "      <td>5.1204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054476</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.3154</td>\n",
       "      <td>4.1574</td>\n",
       "      <td>3.1526</td>\n",
       "      <td>8.9952</td>\n",
       "      <td>3.4898</td>\n",
       "      <td>5.5419</td>\n",
       "      <td>3.5913</td>\n",
       "      <td>3.2972</td>\n",
       "      <td>2.7723</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6189</td>\n",
       "      <td>4.9507</td>\n",
       "      <td>6.7578</td>\n",
       "      <td>7.4825</td>\n",
       "      <td>6.4911</td>\n",
       "      <td>3.4000</td>\n",
       "      <td>8.4330</td>\n",
       "      <td>7.2920</td>\n",
       "      <td>6.4264</td>\n",
       "      <td>5.2261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054477</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.1071</td>\n",
       "      <td>4.0192</td>\n",
       "      <td>2.9848</td>\n",
       "      <td>8.8596</td>\n",
       "      <td>3.3371</td>\n",
       "      <td>5.6346</td>\n",
       "      <td>3.7282</td>\n",
       "      <td>3.2119</td>\n",
       "      <td>3.1112</td>\n",
       "      <td>...</td>\n",
       "      <td>3.4474</td>\n",
       "      <td>4.8923</td>\n",
       "      <td>6.2940</td>\n",
       "      <td>7.0948</td>\n",
       "      <td>6.7106</td>\n",
       "      <td>3.4797</td>\n",
       "      <td>9.5951</td>\n",
       "      <td>7.1470</td>\n",
       "      <td>6.5815</td>\n",
       "      <td>4.9348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054479</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.8205</td>\n",
       "      <td>3.2674</td>\n",
       "      <td>2.7552</td>\n",
       "      <td>8.4401</td>\n",
       "      <td>3.0361</td>\n",
       "      <td>5.3265</td>\n",
       "      <td>3.3717</td>\n",
       "      <td>2.7894</td>\n",
       "      <td>3.1582</td>\n",
       "      <td>...</td>\n",
       "      <td>3.8426</td>\n",
       "      <td>4.2719</td>\n",
       "      <td>6.1915</td>\n",
       "      <td>6.7497</td>\n",
       "      <td>6.6675</td>\n",
       "      <td>3.2020</td>\n",
       "      <td>9.7000</td>\n",
       "      <td>6.7836</td>\n",
       "      <td>6.7070</td>\n",
       "      <td>5.2438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054480</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.9297</td>\n",
       "      <td>3.4582</td>\n",
       "      <td>2.8052</td>\n",
       "      <td>7.9580</td>\n",
       "      <td>3.2262</td>\n",
       "      <td>5.3643</td>\n",
       "      <td>3.4381</td>\n",
       "      <td>3.3667</td>\n",
       "      <td>2.8098</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5798</td>\n",
       "      <td>4.4320</td>\n",
       "      <td>6.1925</td>\n",
       "      <td>7.3337</td>\n",
       "      <td>7.0059</td>\n",
       "      <td>3.3508</td>\n",
       "      <td>9.2442</td>\n",
       "      <td>7.3105</td>\n",
       "      <td>6.9813</td>\n",
       "      <td>5.3024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054483</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.1294</td>\n",
       "      <td>3.9094</td>\n",
       "      <td>2.9174</td>\n",
       "      <td>8.8233</td>\n",
       "      <td>3.3229</td>\n",
       "      <td>5.5326</td>\n",
       "      <td>3.7125</td>\n",
       "      <td>2.8848</td>\n",
       "      <td>3.3188</td>\n",
       "      <td>...</td>\n",
       "      <td>3.4678</td>\n",
       "      <td>4.5581</td>\n",
       "      <td>6.0581</td>\n",
       "      <td>7.3145</td>\n",
       "      <td>6.9052</td>\n",
       "      <td>3.5155</td>\n",
       "      <td>9.3226</td>\n",
       "      <td>7.0620</td>\n",
       "      <td>6.8770</td>\n",
       "      <td>4.9623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054484</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.6635</td>\n",
       "      <td>4.3272</td>\n",
       "      <td>3.2449</td>\n",
       "      <td>8.7281</td>\n",
       "      <td>3.6602</td>\n",
       "      <td>5.8144</td>\n",
       "      <td>4.1438</td>\n",
       "      <td>3.6652</td>\n",
       "      <td>3.4804</td>\n",
       "      <td>...</td>\n",
       "      <td>3.4178</td>\n",
       "      <td>4.6666</td>\n",
       "      <td>6.5508</td>\n",
       "      <td>7.3763</td>\n",
       "      <td>6.4566</td>\n",
       "      <td>3.9737</td>\n",
       "      <td>7.7429</td>\n",
       "      <td>6.9145</td>\n",
       "      <td>6.2181</td>\n",
       "      <td>4.7005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054485</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.8798</td>\n",
       "      <td>3.5844</td>\n",
       "      <td>2.8676</td>\n",
       "      <td>8.2077</td>\n",
       "      <td>3.2373</td>\n",
       "      <td>5.7064</td>\n",
       "      <td>3.4613</td>\n",
       "      <td>2.8435</td>\n",
       "      <td>2.9355</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5886</td>\n",
       "      <td>4.3702</td>\n",
       "      <td>6.0213</td>\n",
       "      <td>7.1321</td>\n",
       "      <td>6.9038</td>\n",
       "      <td>3.2696</td>\n",
       "      <td>9.3012</td>\n",
       "      <td>7.1332</td>\n",
       "      <td>6.7474</td>\n",
       "      <td>5.2812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054487</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5.6396</td>\n",
       "      <td>3.6012</td>\n",
       "      <td>2.7803</td>\n",
       "      <td>8.6148</td>\n",
       "      <td>3.0140</td>\n",
       "      <td>5.6817</td>\n",
       "      <td>3.3996</td>\n",
       "      <td>2.8364</td>\n",
       "      <td>3.0915</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6726</td>\n",
       "      <td>4.4691</td>\n",
       "      <td>6.4462</td>\n",
       "      <td>7.1664</td>\n",
       "      <td>6.9852</td>\n",
       "      <td>3.1801</td>\n",
       "      <td>9.5348</td>\n",
       "      <td>7.2281</td>\n",
       "      <td>6.7670</td>\n",
       "      <td>5.1422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054490</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.4385</td>\n",
       "      <td>7.1722</td>\n",
       "      <td>3.2204</td>\n",
       "      <td>5.3201</td>\n",
       "      <td>4.0520</td>\n",
       "      <td>7.6206</td>\n",
       "      <td>3.7007</td>\n",
       "      <td>4.4598</td>\n",
       "      <td>3.4722</td>\n",
       "      <td>...</td>\n",
       "      <td>6.8044</td>\n",
       "      <td>6.6496</td>\n",
       "      <td>7.2479</td>\n",
       "      <td>8.5252</td>\n",
       "      <td>8.0012</td>\n",
       "      <td>4.5764</td>\n",
       "      <td>9.7288</td>\n",
       "      <td>10.6232</td>\n",
       "      <td>8.3687</td>\n",
       "      <td>8.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054491</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0004</td>\n",
       "      <td>7.0466</td>\n",
       "      <td>3.1001</td>\n",
       "      <td>6.5771</td>\n",
       "      <td>3.7587</td>\n",
       "      <td>6.7833</td>\n",
       "      <td>3.7432</td>\n",
       "      <td>4.4864</td>\n",
       "      <td>3.4312</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1252</td>\n",
       "      <td>6.0701</td>\n",
       "      <td>7.3167</td>\n",
       "      <td>8.3906</td>\n",
       "      <td>8.1439</td>\n",
       "      <td>4.3133</td>\n",
       "      <td>9.2621</td>\n",
       "      <td>9.9298</td>\n",
       "      <td>8.2939</td>\n",
       "      <td>8.4370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054492</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.1297</td>\n",
       "      <td>7.3300</td>\n",
       "      <td>3.2616</td>\n",
       "      <td>4.8996</td>\n",
       "      <td>3.7891</td>\n",
       "      <td>7.5249</td>\n",
       "      <td>4.0087</td>\n",
       "      <td>4.7360</td>\n",
       "      <td>3.6205</td>\n",
       "      <td>...</td>\n",
       "      <td>7.6245</td>\n",
       "      <td>7.5046</td>\n",
       "      <td>7.0205</td>\n",
       "      <td>8.1047</td>\n",
       "      <td>7.8983</td>\n",
       "      <td>4.1094</td>\n",
       "      <td>9.5219</td>\n",
       "      <td>10.4447</td>\n",
       "      <td>8.7851</td>\n",
       "      <td>8.1577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054493</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.2450</td>\n",
       "      <td>7.0268</td>\n",
       "      <td>3.2589</td>\n",
       "      <td>5.6858</td>\n",
       "      <td>3.7584</td>\n",
       "      <td>7.5707</td>\n",
       "      <td>4.0084</td>\n",
       "      <td>4.2058</td>\n",
       "      <td>3.4574</td>\n",
       "      <td>...</td>\n",
       "      <td>7.4832</td>\n",
       "      <td>7.0871</td>\n",
       "      <td>7.4019</td>\n",
       "      <td>8.2968</td>\n",
       "      <td>8.0049</td>\n",
       "      <td>4.5146</td>\n",
       "      <td>8.9699</td>\n",
       "      <td>10.0450</td>\n",
       "      <td>8.2137</td>\n",
       "      <td>7.8236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054494</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.1468</td>\n",
       "      <td>7.0425</td>\n",
       "      <td>3.1183</td>\n",
       "      <td>6.8993</td>\n",
       "      <td>3.7080</td>\n",
       "      <td>7.3469</td>\n",
       "      <td>4.0332</td>\n",
       "      <td>3.9393</td>\n",
       "      <td>3.2263</td>\n",
       "      <td>...</td>\n",
       "      <td>7.4996</td>\n",
       "      <td>7.0295</td>\n",
       "      <td>7.0373</td>\n",
       "      <td>8.0565</td>\n",
       "      <td>8.0106</td>\n",
       "      <td>4.4156</td>\n",
       "      <td>9.2271</td>\n",
       "      <td>10.4999</td>\n",
       "      <td>8.3953</td>\n",
       "      <td>8.1731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054496</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.3677</td>\n",
       "      <td>6.9973</td>\n",
       "      <td>3.2747</td>\n",
       "      <td>7.3920</td>\n",
       "      <td>4.1103</td>\n",
       "      <td>7.2068</td>\n",
       "      <td>4.0988</td>\n",
       "      <td>4.8371</td>\n",
       "      <td>3.4701</td>\n",
       "      <td>...</td>\n",
       "      <td>7.3735</td>\n",
       "      <td>6.6522</td>\n",
       "      <td>7.4772</td>\n",
       "      <td>8.4933</td>\n",
       "      <td>7.8751</td>\n",
       "      <td>4.2377</td>\n",
       "      <td>9.5495</td>\n",
       "      <td>10.3390</td>\n",
       "      <td>8.0165</td>\n",
       "      <td>8.1466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054497</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.4013</td>\n",
       "      <td>6.9534</td>\n",
       "      <td>3.3139</td>\n",
       "      <td>6.9890</td>\n",
       "      <td>3.9745</td>\n",
       "      <td>8.0541</td>\n",
       "      <td>4.2043</td>\n",
       "      <td>4.2463</td>\n",
       "      <td>3.5869</td>\n",
       "      <td>...</td>\n",
       "      <td>6.8589</td>\n",
       "      <td>6.8268</td>\n",
       "      <td>7.4688</td>\n",
       "      <td>8.6030</td>\n",
       "      <td>8.1382</td>\n",
       "      <td>4.4298</td>\n",
       "      <td>9.5942</td>\n",
       "      <td>10.5005</td>\n",
       "      <td>8.1637</td>\n",
       "      <td>8.0366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054498</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.3875</td>\n",
       "      <td>7.0710</td>\n",
       "      <td>3.4451</td>\n",
       "      <td>5.8473</td>\n",
       "      <td>3.9496</td>\n",
       "      <td>7.2179</td>\n",
       "      <td>4.2349</td>\n",
       "      <td>4.6678</td>\n",
       "      <td>3.5877</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0902</td>\n",
       "      <td>6.0270</td>\n",
       "      <td>7.6471</td>\n",
       "      <td>8.4840</td>\n",
       "      <td>8.2386</td>\n",
       "      <td>4.5878</td>\n",
       "      <td>9.3331</td>\n",
       "      <td>9.6730</td>\n",
       "      <td>8.1882</td>\n",
       "      <td>7.9059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054499</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.2522</td>\n",
       "      <td>6.8242</td>\n",
       "      <td>3.1718</td>\n",
       "      <td>6.4205</td>\n",
       "      <td>3.9631</td>\n",
       "      <td>7.4694</td>\n",
       "      <td>3.9543</td>\n",
       "      <td>3.8568</td>\n",
       "      <td>3.6741</td>\n",
       "      <td>...</td>\n",
       "      <td>6.7595</td>\n",
       "      <td>6.6548</td>\n",
       "      <td>7.2312</td>\n",
       "      <td>8.3601</td>\n",
       "      <td>8.0991</td>\n",
       "      <td>4.0764</td>\n",
       "      <td>9.0859</td>\n",
       "      <td>10.0572</td>\n",
       "      <td>8.3218</td>\n",
       "      <td>8.2084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054500</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.3063</td>\n",
       "      <td>6.9274</td>\n",
       "      <td>3.1437</td>\n",
       "      <td>6.2682</td>\n",
       "      <td>3.8923</td>\n",
       "      <td>7.5651</td>\n",
       "      <td>4.0858</td>\n",
       "      <td>4.6939</td>\n",
       "      <td>3.4536</td>\n",
       "      <td>...</td>\n",
       "      <td>6.8811</td>\n",
       "      <td>6.5652</td>\n",
       "      <td>7.3402</td>\n",
       "      <td>8.5191</td>\n",
       "      <td>8.1215</td>\n",
       "      <td>4.6576</td>\n",
       "      <td>9.3023</td>\n",
       "      <td>10.3978</td>\n",
       "      <td>8.3739</td>\n",
       "      <td>8.1474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054502</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.5244</td>\n",
       "      <td>7.1620</td>\n",
       "      <td>3.4615</td>\n",
       "      <td>10.4299</td>\n",
       "      <td>3.9915</td>\n",
       "      <td>7.4478</td>\n",
       "      <td>4.1612</td>\n",
       "      <td>4.5842</td>\n",
       "      <td>3.7893</td>\n",
       "      <td>...</td>\n",
       "      <td>6.9901</td>\n",
       "      <td>7.2414</td>\n",
       "      <td>7.4043</td>\n",
       "      <td>8.3254</td>\n",
       "      <td>7.8031</td>\n",
       "      <td>4.6602</td>\n",
       "      <td>9.2111</td>\n",
       "      <td>10.4286</td>\n",
       "      <td>8.5387</td>\n",
       "      <td>7.5695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054503</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.3643</td>\n",
       "      <td>7.0130</td>\n",
       "      <td>3.2555</td>\n",
       "      <td>5.0785</td>\n",
       "      <td>3.8972</td>\n",
       "      <td>7.5749</td>\n",
       "      <td>3.7441</td>\n",
       "      <td>3.9326</td>\n",
       "      <td>3.4924</td>\n",
       "      <td>...</td>\n",
       "      <td>7.7016</td>\n",
       "      <td>7.4546</td>\n",
       "      <td>7.0112</td>\n",
       "      <td>8.0315</td>\n",
       "      <td>7.8589</td>\n",
       "      <td>5.7572</td>\n",
       "      <td>9.5072</td>\n",
       "      <td>10.8868</td>\n",
       "      <td>8.5115</td>\n",
       "      <td>8.1701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054505</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.3470</td>\n",
       "      <td>6.8703</td>\n",
       "      <td>3.1797</td>\n",
       "      <td>8.4455</td>\n",
       "      <td>3.7672</td>\n",
       "      <td>6.5622</td>\n",
       "      <td>3.7452</td>\n",
       "      <td>4.5728</td>\n",
       "      <td>3.4647</td>\n",
       "      <td>...</td>\n",
       "      <td>7.7980</td>\n",
       "      <td>7.6082</td>\n",
       "      <td>7.0637</td>\n",
       "      <td>8.3682</td>\n",
       "      <td>7.8773</td>\n",
       "      <td>4.7218</td>\n",
       "      <td>9.6251</td>\n",
       "      <td>10.9998</td>\n",
       "      <td>8.1910</td>\n",
       "      <td>7.9798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054506</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.3858</td>\n",
       "      <td>6.5639</td>\n",
       "      <td>3.2482</td>\n",
       "      <td>6.4056</td>\n",
       "      <td>3.9695</td>\n",
       "      <td>8.1602</td>\n",
       "      <td>3.8948</td>\n",
       "      <td>4.8346</td>\n",
       "      <td>3.4946</td>\n",
       "      <td>...</td>\n",
       "      <td>6.9912</td>\n",
       "      <td>6.6165</td>\n",
       "      <td>7.2670</td>\n",
       "      <td>8.4671</td>\n",
       "      <td>8.0801</td>\n",
       "      <td>4.8042</td>\n",
       "      <td>9.4461</td>\n",
       "      <td>10.7922</td>\n",
       "      <td>8.3599</td>\n",
       "      <td>7.7927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054507</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.2936</td>\n",
       "      <td>7.2063</td>\n",
       "      <td>3.1539</td>\n",
       "      <td>5.0722</td>\n",
       "      <td>3.8859</td>\n",
       "      <td>7.1595</td>\n",
       "      <td>4.0312</td>\n",
       "      <td>4.1642</td>\n",
       "      <td>3.4708</td>\n",
       "      <td>...</td>\n",
       "      <td>7.5825</td>\n",
       "      <td>7.1933</td>\n",
       "      <td>7.1169</td>\n",
       "      <td>8.1924</td>\n",
       "      <td>8.0482</td>\n",
       "      <td>4.0250</td>\n",
       "      <td>9.4394</td>\n",
       "      <td>10.5329</td>\n",
       "      <td>8.4375</td>\n",
       "      <td>7.9860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054508</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.3586</td>\n",
       "      <td>6.9356</td>\n",
       "      <td>3.3267</td>\n",
       "      <td>7.2323</td>\n",
       "      <td>3.8911</td>\n",
       "      <td>7.5459</td>\n",
       "      <td>3.8832</td>\n",
       "      <td>3.6907</td>\n",
       "      <td>3.5198</td>\n",
       "      <td>...</td>\n",
       "      <td>7.3888</td>\n",
       "      <td>7.2046</td>\n",
       "      <td>7.2724</td>\n",
       "      <td>8.3439</td>\n",
       "      <td>8.0633</td>\n",
       "      <td>4.4530</td>\n",
       "      <td>9.2585</td>\n",
       "      <td>10.4911</td>\n",
       "      <td>8.4334</td>\n",
       "      <td>7.9013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054509</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.5003</td>\n",
       "      <td>6.7637</td>\n",
       "      <td>3.2850</td>\n",
       "      <td>6.9800</td>\n",
       "      <td>4.0563</td>\n",
       "      <td>7.0698</td>\n",
       "      <td>4.1589</td>\n",
       "      <td>3.6182</td>\n",
       "      <td>3.4340</td>\n",
       "      <td>...</td>\n",
       "      <td>6.9541</td>\n",
       "      <td>6.4660</td>\n",
       "      <td>7.4200</td>\n",
       "      <td>8.6939</td>\n",
       "      <td>8.0615</td>\n",
       "      <td>4.5717</td>\n",
       "      <td>9.0012</td>\n",
       "      <td>10.2403</td>\n",
       "      <td>8.3261</td>\n",
       "      <td>7.7559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054511</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.2938</td>\n",
       "      <td>6.9173</td>\n",
       "      <td>3.2007</td>\n",
       "      <td>5.0774</td>\n",
       "      <td>3.9923</td>\n",
       "      <td>7.1605</td>\n",
       "      <td>3.8314</td>\n",
       "      <td>3.9822</td>\n",
       "      <td>3.1061</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1835</td>\n",
       "      <td>7.1075</td>\n",
       "      <td>7.1243</td>\n",
       "      <td>8.3166</td>\n",
       "      <td>8.1081</td>\n",
       "      <td>4.7296</td>\n",
       "      <td>9.3080</td>\n",
       "      <td>10.5929</td>\n",
       "      <td>8.3949</td>\n",
       "      <td>8.0305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054512</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.2880</td>\n",
       "      <td>6.9877</td>\n",
       "      <td>3.3306</td>\n",
       "      <td>5.0126</td>\n",
       "      <td>3.7533</td>\n",
       "      <td>7.6316</td>\n",
       "      <td>4.0710</td>\n",
       "      <td>4.4998</td>\n",
       "      <td>3.4946</td>\n",
       "      <td>...</td>\n",
       "      <td>7.2733</td>\n",
       "      <td>6.7483</td>\n",
       "      <td>7.0236</td>\n",
       "      <td>8.3423</td>\n",
       "      <td>7.7346</td>\n",
       "      <td>4.4920</td>\n",
       "      <td>9.1942</td>\n",
       "      <td>10.4838</td>\n",
       "      <td>8.2926</td>\n",
       "      <td>7.8095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1054513</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.3009</td>\n",
       "      <td>6.8485</td>\n",
       "      <td>3.1708</td>\n",
       "      <td>10.3406</td>\n",
       "      <td>4.0087</td>\n",
       "      <td>7.9692</td>\n",
       "      <td>4.3088</td>\n",
       "      <td>4.6175</td>\n",
       "      <td>3.5245</td>\n",
       "      <td>...</td>\n",
       "      <td>7.1821</td>\n",
       "      <td>6.9740</td>\n",
       "      <td>7.3127</td>\n",
       "      <td>8.4597</td>\n",
       "      <td>7.8855</td>\n",
       "      <td>4.3177</td>\n",
       "      <td>9.4302</td>\n",
       "      <td>10.8218</td>\n",
       "      <td>8.2379</td>\n",
       "      <td>7.6898</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43 rows × 17545 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Mitochondrial Disorders    A1BG  A1BG-AS1    A1CF      A2M  \\\n",
       "GSM1054461                      0.0  6.1013    4.1155  3.0490   9.5803   \n",
       "GSM1054462                      1.0  5.8975    3.8231  2.8329   8.8574   \n",
       "GSM1054463                      1.0  6.1725    4.4809  3.0462   8.6237   \n",
       "GSM1054464                      1.0  5.9452    3.8832  3.0998   8.7689   \n",
       "GSM1054465                      1.0  5.8125    3.5292  2.8448   9.6166   \n",
       "GSM1054466                      0.0  6.2270    4.0760  2.9790   9.5560   \n",
       "GSM1054467                      0.0  6.6265    4.2817  3.1286   9.5490   \n",
       "GSM1054468                      0.0  6.1640    4.0797  3.2239   8.8824   \n",
       "GSM1054469                      0.0  6.2533    3.7546  2.9757   8.9636   \n",
       "GSM1054470                      1.0  5.9966    3.7637  2.8277   8.7677   \n",
       "GSM1054471                      1.0  5.6345    3.6525  2.8254   9.5466   \n",
       "GSM1054472                      0.0  5.9679    3.8574  3.0826   9.0008   \n",
       "GSM1054473                      1.0  5.6092    3.4124  2.7425   9.7295   \n",
       "GSM1054474                      0.0  5.9416    3.9226  2.8950   9.4086   \n",
       "GSM1054475                      1.0  5.8829    3.5686  2.9108   8.9915   \n",
       "GSM1054476                      1.0  6.3154    4.1574  3.1526   8.9952   \n",
       "GSM1054477                      1.0  6.1071    4.0192  2.9848   8.8596   \n",
       "GSM1054479                      1.0  5.8205    3.2674  2.7552   8.4401   \n",
       "GSM1054480                      1.0  5.9297    3.4582  2.8052   7.9580   \n",
       "GSM1054483                      0.0  6.1294    3.9094  2.9174   8.8233   \n",
       "GSM1054484                      0.0  6.6635    4.3272  3.2449   8.7281   \n",
       "GSM1054485                      0.0  5.8798    3.5844  2.8676   8.2077   \n",
       "GSM1054487                      1.0  5.6396    3.6012  2.7803   8.6148   \n",
       "GSM1054490                      0.0  7.4385    7.1722  3.2204   5.3201   \n",
       "GSM1054491                      0.0  7.0004    7.0466  3.1001   6.5771   \n",
       "GSM1054492                      0.0  7.1297    7.3300  3.2616   4.8996   \n",
       "GSM1054493                      1.0  7.2450    7.0268  3.2589   5.6858   \n",
       "GSM1054494                      0.0  7.1468    7.0425  3.1183   6.8993   \n",
       "GSM1054496                      1.0  7.3677    6.9973  3.2747   7.3920   \n",
       "GSM1054497                      1.0  7.4013    6.9534  3.3139   6.9890   \n",
       "GSM1054498                      1.0  7.3875    7.0710  3.4451   5.8473   \n",
       "GSM1054499                      1.0  7.2522    6.8242  3.1718   6.4205   \n",
       "GSM1054500                      0.0  7.3063    6.9274  3.1437   6.2682   \n",
       "GSM1054502                      1.0  7.5244    7.1620  3.4615  10.4299   \n",
       "GSM1054503                      0.0  7.3643    7.0130  3.2555   5.0785   \n",
       "GSM1054505                      0.0  7.3470    6.8703  3.1797   8.4455   \n",
       "GSM1054506                      1.0  7.3858    6.5639  3.2482   6.4056   \n",
       "GSM1054507                      0.0  7.2936    7.2063  3.1539   5.0722   \n",
       "GSM1054508                      1.0  7.3586    6.9356  3.3267   7.2323   \n",
       "GSM1054509                      1.0  7.5003    6.7637  3.2850   6.9800   \n",
       "GSM1054511                      1.0  7.2938    6.9173  3.2007   5.0774   \n",
       "GSM1054512                      1.0  7.2880    6.9877  3.3306   5.0126   \n",
       "GSM1054513                      1.0  7.3009    6.8485  3.1708  10.3406   \n",
       "\n",
       "             A2ML1  A4GALT   A4GNT    AA06    AAA1  ...  ZWILCH   ZWINT  \\\n",
       "GSM1054461  3.2303  5.7970  3.4187  3.6406  3.0490  ...  3.6210  4.6627   \n",
       "GSM1054462  3.1281  5.2234  4.0981  2.8983  3.5197  ...  3.6732  4.0082   \n",
       "GSM1054463  3.3229  5.5002  3.6228  4.0165  3.1844  ...  3.7046  4.8828   \n",
       "GSM1054464  3.3928  5.6048  3.4490  3.0753  3.0915  ...  3.3512  4.4825   \n",
       "GSM1054465  3.1351  5.6190  3.5056  3.2483  2.7159  ...  3.8325  4.7506   \n",
       "GSM1054466  3.3229  5.6816  3.4400  3.3479  2.8350  ...  3.8073  4.5413   \n",
       "GSM1054467  3.7033  5.7616  3.8976  3.4092  3.0747  ...  3.5066  4.8491   \n",
       "GSM1054468  3.3818  5.6999  3.7429  3.4481  3.0651  ...  3.5940  4.7115   \n",
       "GSM1054469  3.2558  5.5583  4.3190  3.5288  3.3684  ...  3.5712  4.4865   \n",
       "GSM1054470  3.2793  6.5516  3.4250  3.2469  2.7997  ...  3.3897  4.4323   \n",
       "GSM1054471  2.9381  6.0614  3.2169  2.9694  2.8535  ...  3.6847  4.4523   \n",
       "GSM1054472  3.3229  5.7816  4.2293  3.2270  3.5524  ...  3.5150  4.5171   \n",
       "GSM1054473  2.9885  5.3000  3.5026  2.5168  3.0143  ...  3.5778  4.2257   \n",
       "GSM1054474  3.1365  5.3408  3.4832  3.1897  2.9552  ...  3.7758  4.3914   \n",
       "GSM1054475  3.2265  5.2228  3.4857  3.9896  3.0070  ...  3.5944  4.5104   \n",
       "GSM1054476  3.4898  5.5419  3.5913  3.2972  2.7723  ...  3.6189  4.9507   \n",
       "GSM1054477  3.3371  5.6346  3.7282  3.2119  3.1112  ...  3.4474  4.8923   \n",
       "GSM1054479  3.0361  5.3265  3.3717  2.7894  3.1582  ...  3.8426  4.2719   \n",
       "GSM1054480  3.2262  5.3643  3.4381  3.3667  2.8098  ...  3.5798  4.4320   \n",
       "GSM1054483  3.3229  5.5326  3.7125  2.8848  3.3188  ...  3.4678  4.5581   \n",
       "GSM1054484  3.6602  5.8144  4.1438  3.6652  3.4804  ...  3.4178  4.6666   \n",
       "GSM1054485  3.2373  5.7064  3.4613  2.8435  2.9355  ...  3.5886  4.3702   \n",
       "GSM1054487  3.0140  5.6817  3.3996  2.8364  3.0915  ...  3.6726  4.4691   \n",
       "GSM1054490  4.0520  7.6206  3.7007  4.4598  3.4722  ...  6.8044  6.6496   \n",
       "GSM1054491  3.7587  6.7833  3.7432  4.4864  3.4312  ...  7.1252  6.0701   \n",
       "GSM1054492  3.7891  7.5249  4.0087  4.7360  3.6205  ...  7.6245  7.5046   \n",
       "GSM1054493  3.7584  7.5707  4.0084  4.2058  3.4574  ...  7.4832  7.0871   \n",
       "GSM1054494  3.7080  7.3469  4.0332  3.9393  3.2263  ...  7.4996  7.0295   \n",
       "GSM1054496  4.1103  7.2068  4.0988  4.8371  3.4701  ...  7.3735  6.6522   \n",
       "GSM1054497  3.9745  8.0541  4.2043  4.2463  3.5869  ...  6.8589  6.8268   \n",
       "GSM1054498  3.9496  7.2179  4.2349  4.6678  3.5877  ...  7.0902  6.0270   \n",
       "GSM1054499  3.9631  7.4694  3.9543  3.8568  3.6741  ...  6.7595  6.6548   \n",
       "GSM1054500  3.8923  7.5651  4.0858  4.6939  3.4536  ...  6.8811  6.5652   \n",
       "GSM1054502  3.9915  7.4478  4.1612  4.5842  3.7893  ...  6.9901  7.2414   \n",
       "GSM1054503  3.8972  7.5749  3.7441  3.9326  3.4924  ...  7.7016  7.4546   \n",
       "GSM1054505  3.7672  6.5622  3.7452  4.5728  3.4647  ...  7.7980  7.6082   \n",
       "GSM1054506  3.9695  8.1602  3.8948  4.8346  3.4946  ...  6.9912  6.6165   \n",
       "GSM1054507  3.8859  7.1595  4.0312  4.1642  3.4708  ...  7.5825  7.1933   \n",
       "GSM1054508  3.8911  7.5459  3.8832  3.6907  3.5198  ...  7.3888  7.2046   \n",
       "GSM1054509  4.0563  7.0698  4.1589  3.6182  3.4340  ...  6.9541  6.4660   \n",
       "GSM1054511  3.9923  7.1605  3.8314  3.9822  3.1061  ...  7.1835  7.1075   \n",
       "GSM1054512  3.7533  7.6316  4.0710  4.4998  3.4946  ...  7.2733  6.7483   \n",
       "GSM1054513  4.0087  7.9692  4.3088  4.6175  3.5245  ...  7.1821  6.9740   \n",
       "\n",
       "              ZXDA    ZXDB    ZXDC  ZYG11A   ZYG11B      ZYX   ZZEF1    ZZZ3  \n",
       "GSM1054461  6.7564  6.9139  6.4468  3.2158   9.0668   7.1701  6.3858  5.2251  \n",
       "GSM1054462  6.3447  7.2577  6.3363  3.7873   9.4642   6.9826  6.5421  5.3963  \n",
       "GSM1054463  6.4101  6.6484  6.4234  3.4058   9.3095   6.9313  6.7484  4.8275  \n",
       "GSM1054464  6.4230  7.1715  6.4230  3.3421   9.4767   7.3664  6.2827  5.0468  \n",
       "GSM1054465  7.4742  7.4161  6.2950  3.3949   8.9353   7.2823  6.6528  5.0241  \n",
       "GSM1054466  6.7705  7.3081  6.4915  3.2535   8.6806   6.6743  6.2846  5.1010  \n",
       "GSM1054467  6.7513  7.0802  6.5757  3.6718   8.8417   6.9897  6.1841  4.8599  \n",
       "GSM1054468  6.8967  7.3867  6.7240  3.5500   8.3534   7.2487  6.1643  4.9404  \n",
       "GSM1054469  6.9326  7.1493  6.7174  3.4669   8.3751   7.2203  6.1597  5.0926  \n",
       "GSM1054470  5.6660  6.7189  7.6097  3.2070   8.3136   8.2768  6.7696  5.2231  \n",
       "GSM1054471  5.8703  6.9296  7.5199  2.9847   8.2782   8.4515  7.5613  5.2294  \n",
       "GSM1054472  6.7663  7.2602  6.5348  3.7670   9.2247   7.3436  6.4235  4.9750  \n",
       "GSM1054473  6.6533  6.9227  6.7628  3.2882  10.1225   7.1470  6.8428  5.4319  \n",
       "GSM1054474  6.3239  7.0742  6.7333  3.4356   9.6568   7.2235  7.2606  4.7891  \n",
       "GSM1054475  6.2869  6.9581  6.8440  3.2626  10.0019   7.1317  7.2185  5.1204  \n",
       "GSM1054476  6.7578  7.4825  6.4911  3.4000   8.4330   7.2920  6.4264  5.2261  \n",
       "GSM1054477  6.2940  7.0948  6.7106  3.4797   9.5951   7.1470  6.5815  4.9348  \n",
       "GSM1054479  6.1915  6.7497  6.6675  3.2020   9.7000   6.7836  6.7070  5.2438  \n",
       "GSM1054480  6.1925  7.3337  7.0059  3.3508   9.2442   7.3105  6.9813  5.3024  \n",
       "GSM1054483  6.0581  7.3145  6.9052  3.5155   9.3226   7.0620  6.8770  4.9623  \n",
       "GSM1054484  6.5508  7.3763  6.4566  3.9737   7.7429   6.9145  6.2181  4.7005  \n",
       "GSM1054485  6.0213  7.1321  6.9038  3.2696   9.3012   7.1332  6.7474  5.2812  \n",
       "GSM1054487  6.4462  7.1664  6.9852  3.1801   9.5348   7.2281  6.7670  5.1422  \n",
       "GSM1054490  7.2479  8.5252  8.0012  4.5764   9.7288  10.6232  8.3687  8.0760  \n",
       "GSM1054491  7.3167  8.3906  8.1439  4.3133   9.2621   9.9298  8.2939  8.4370  \n",
       "GSM1054492  7.0205  8.1047  7.8983  4.1094   9.5219  10.4447  8.7851  8.1577  \n",
       "GSM1054493  7.4019  8.2968  8.0049  4.5146   8.9699  10.0450  8.2137  7.8236  \n",
       "GSM1054494  7.0373  8.0565  8.0106  4.4156   9.2271  10.4999  8.3953  8.1731  \n",
       "GSM1054496  7.4772  8.4933  7.8751  4.2377   9.5495  10.3390  8.0165  8.1466  \n",
       "GSM1054497  7.4688  8.6030  8.1382  4.4298   9.5942  10.5005  8.1637  8.0366  \n",
       "GSM1054498  7.6471  8.4840  8.2386  4.5878   9.3331   9.6730  8.1882  7.9059  \n",
       "GSM1054499  7.2312  8.3601  8.0991  4.0764   9.0859  10.0572  8.3218  8.2084  \n",
       "GSM1054500  7.3402  8.5191  8.1215  4.6576   9.3023  10.3978  8.3739  8.1474  \n",
       "GSM1054502  7.4043  8.3254  7.8031  4.6602   9.2111  10.4286  8.5387  7.5695  \n",
       "GSM1054503  7.0112  8.0315  7.8589  5.7572   9.5072  10.8868  8.5115  8.1701  \n",
       "GSM1054505  7.0637  8.3682  7.8773  4.7218   9.6251  10.9998  8.1910  7.9798  \n",
       "GSM1054506  7.2670  8.4671  8.0801  4.8042   9.4461  10.7922  8.3599  7.7927  \n",
       "GSM1054507  7.1169  8.1924  8.0482  4.0250   9.4394  10.5329  8.4375  7.9860  \n",
       "GSM1054508  7.2724  8.3439  8.0633  4.4530   9.2585  10.4911  8.4334  7.9013  \n",
       "GSM1054509  7.4200  8.6939  8.0615  4.5717   9.0012  10.2403  8.3261  7.7559  \n",
       "GSM1054511  7.1243  8.3166  8.1081  4.7296   9.3080  10.5929  8.3949  8.0305  \n",
       "GSM1054512  7.0236  8.3423  7.7346  4.4920   9.1942  10.4838  8.2926  7.8095  \n",
       "GSM1054513  7.3127  8.4597  7.8855  4.3177   9.4302  10.8218  8.2379  7.6898  \n",
       "\n",
       "[43 rows x 17545 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 43 samples.\n",
      "For the feature 'Mitochondrial Disorders', the least common label is '0.0' with 18 occurrences. This represents 41.86% of the dataset.\n",
      "The distribution of the feature 'Mitochondrial Disorders' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output2\\Jiayi\\Fibromyalgia\\cohort_info.json\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'merged_data' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[12], line 7\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m      6\u001b[0m     save_cohort_info(cohort, JSON_PATH, is_available)\n\u001b[1;32m----> 7\u001b[0m \u001b[43mmerged_data\u001b[49m\u001b[38;5;241m.\u001b[39mhead()\n\u001b[0;32m      8\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_trait_biased:\n\u001b[0;32m      9\u001b[0m     merged_data\u001b[38;5;241m.\u001b[39mto_csv(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(OUTPUT_DIR, cohort \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m), index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'merged_data' is not defined"
     ]
    }
   ],
   "source": [
    "is_available = False\n",
    "\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
