{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = \"Paget's Disease of bone\"\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE7545', 'GSE7849']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Paget_s-Disease-of-bone\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Paget_s-Disease-of-bone\\\\GSE7545\\\\GSE7545_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Paget_s-Disease-of-bone\\\\GSE7545\\\\GSE7545-GPL3718_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE7545\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Paget_s-Disease-of-bone\\\\GSE7849\\\\GSE7849_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Paget_s-Disease-of-bone\\\\GSE7849\\\\GSE7849_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The clinical data is biased\n",
    "cohort = accession_num = \"GSE7849\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Age-Specific Differences of Oncogenic Pathway Deregulation Seen in Human Breast Tumors\"\n",
      "!Series_summary\t\"Breast cancer arising in young women has a poorer prognosis, is less likely to be hormone sensitive, and represents a particularly challenging clinical entity.  The biology driving the aggressive nature of breast cancer arising in young women has yet to be defined.\"\n",
      "!Series_summary\t\"Among 784 patients with early stage breast cancer, using prospectively-defined, age-specific cohorts (young <= 45 years; older >= 65 years), 411 eligible patients (n = 200 < 45 years; n = 211 >= 65 years) with clinically-annotated Affymetrix microarray data were identified. Gene set enrichment analyses, signatures of oncogenic pathway deregulation and predictors of chemotherapy sensitivity were evaluated within the two age-defined cohorts.\"\n",
      "!Series_summary\t\"In comparing deregulation of oncogenic pathways between age groups, a statistically higher probability of PI3K (p = 0.006) and Myc (p = 0.03) pathway deregulation was observed in the tumors of younger women.  When evaluating unique patterns of pathway deregulation, a low probability of Src and E2F deregulation in tumors of younger women, concurrent with activation of PI3K, Myc, and beta-catenin, conferred a worse prognosis (HR = 4.15; p = 0.008).  In contrast, a higher probability of Src and E2F pathway activation in tumors of older women, concurrent low probability of PI3K, Myc and beta-catenin deregulation, was associated with a poorer outcome (HR = 2.7; p = 0.006). Similar pathway differences were identified using gene set enrichment analysis. Importantly, in multivariate analyses including clinico-pathologic variables, genomic clusters of pathway deregulation were identified to be independent predictors of disease-free survival.  Finally, a significant relationship (p = 0.02) between anthracycline sensitivity and genomic clusters was observed among women aged >= 65 years.\"\n",
      "!Series_summary\t\"\"\n",
      "!Series_summary\t\"Submitters do not have approval to publish the .CEL files\"\n",
      "!Series_summary\t\"Keywords: Retrospective analyses\"\n",
      "!Series_overall_design\t\"n=78\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM190415</th>\n",
       "      <th>GSM190416</th>\n",
       "      <th>GSM190417</th>\n",
       "      <th>GSM190418</th>\n",
       "      <th>GSM190419</th>\n",
       "      <th>GSM190420</th>\n",
       "      <th>GSM190421</th>\n",
       "      <th>GSM190422</th>\n",
       "      <th>GSM190423</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM190483</th>\n",
       "      <th>GSM190484</th>\n",
       "      <th>GSM190485</th>\n",
       "      <th>GSM190486</th>\n",
       "      <th>GSM190487</th>\n",
       "      <th>GSM190488</th>\n",
       "      <th>GSM190489</th>\n",
       "      <th>GSM190490</th>\n",
       "      <th>GSM190491</th>\n",
       "      <th>GSM190492</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>EA_Designation = 1</td>\n",
       "      <td>EA_Designation = 2</td>\n",
       "      <td>EA_Designation = 3</td>\n",
       "      <td>EA_Designation = 4</td>\n",
       "      <td>EA_Designation = 5</td>\n",
       "      <td>EA_Designation = 6</td>\n",
       "      <td>EA_Designation = 7</td>\n",
       "      <td>EA_Designation = 8</td>\n",
       "      <td>EA_Designation = 9</td>\n",
       "      <td>...</td>\n",
       "      <td>EA_Designation = 80</td>\n",
       "      <td>EA_Designation = 81</td>\n",
       "      <td>EA_Designation = 82</td>\n",
       "      <td>EA_Designation = 83</td>\n",
       "      <td>EA_Designation = 85</td>\n",
       "      <td>EA_Designation = 86</td>\n",
       "      <td>EA_Designation = 87</td>\n",
       "      <td>EA_Designation = 88</td>\n",
       "      <td>EA_Designation = 89</td>\n",
       "      <td>EA_Designation = 90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>%_Tumor = 80</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 70</td>\n",
       "      <td>%_Tumor = 50</td>\n",
       "      <td>%_Tumor = 50</td>\n",
       "      <td>%_Tumor = 90</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 50</td>\n",
       "      <td>%_Tumor = 80</td>\n",
       "      <td>...</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 50</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 70</td>\n",
       "      <td>%_Tumor = 50</td>\n",
       "      <td>%_Tumor = 90</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "      <td>%_Tumor = 60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = Black</td>\n",
       "      <td>Race = Asian</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>...</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = Black</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = White</td>\n",
       "      <td>Race = Black</td>\n",
       "      <td>Race = Am Indian</td>\n",
       "      <td>Race = White</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Infiltrating Duct &amp; Lobualar Carc...</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>...</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Infiltrating Duct &amp; Lobualar Carc...</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Paget Disease &amp; Infiltrating Duct...</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "      <td>Histo_Desc = Ductal Carcinoma, NOS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                           GSM190415  \\\n",
       "0  !Sample_characteristics_ch1                               Tumor   \n",
       "1  !Sample_characteristics_ch1                  EA_Designation = 1   \n",
       "2  !Sample_characteristics_ch1                        %_Tumor = 80   \n",
       "3  !Sample_characteristics_ch1                        Race = White   \n",
       "4  !Sample_characteristics_ch1  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                            GSM190416                           GSM190417  \\\n",
       "0                               Tumor                               Tumor   \n",
       "1                  EA_Designation = 2                  EA_Designation = 3   \n",
       "2                        %_Tumor = 60                        %_Tumor = 70   \n",
       "3                        Race = White                        Race = White   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                            GSM190418                           GSM190419  \\\n",
       "0                               Tumor                               Tumor   \n",
       "1                  EA_Designation = 4                  EA_Designation = 5   \n",
       "2                        %_Tumor = 50                        %_Tumor = 50   \n",
       "3                        Race = Black                        Race = Asian   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                            GSM190420                           GSM190421  \\\n",
       "0                               Tumor                               Tumor   \n",
       "1                  EA_Designation = 6                  EA_Designation = 7   \n",
       "2                        %_Tumor = 90                        %_Tumor = 60   \n",
       "3                        Race = White                        Race = White   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                                           GSM190422  \\\n",
       "0                                              Tumor   \n",
       "1                                 EA_Designation = 8   \n",
       "2                                       %_Tumor = 50   \n",
       "3                                       Race = White   \n",
       "4  Histo_Desc = Infiltrating Duct & Lobualar Carc...   \n",
       "\n",
       "                            GSM190423  ...  \\\n",
       "0                               Tumor  ...   \n",
       "1                  EA_Designation = 9  ...   \n",
       "2                        %_Tumor = 80  ...   \n",
       "3                        Race = White  ...   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  ...   \n",
       "\n",
       "                            GSM190483                           GSM190484  \\\n",
       "0                               Tumor                               Tumor   \n",
       "1                 EA_Designation = 80                 EA_Designation = 81   \n",
       "2                        %_Tumor = 60                        %_Tumor = 60   \n",
       "3                        Race = White                        Race = White   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                                           GSM190485  \\\n",
       "0                                              Tumor   \n",
       "1                                EA_Designation = 82   \n",
       "2                                       %_Tumor = 50   \n",
       "3                                       Race = Black   \n",
       "4  Histo_Desc = Infiltrating Duct & Lobualar Carc...   \n",
       "\n",
       "                            GSM190486  \\\n",
       "0                               Tumor   \n",
       "1                 EA_Designation = 83   \n",
       "2                        %_Tumor = 60   \n",
       "3                        Race = White   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                                           GSM190487  \\\n",
       "0                                              Tumor   \n",
       "1                                EA_Designation = 85   \n",
       "2                                       %_Tumor = 70   \n",
       "3                                       Race = White   \n",
       "4  Histo_Desc = Paget Disease & Infiltrating Duct...   \n",
       "\n",
       "                            GSM190488                    GSM190489  \\\n",
       "0                               Tumor                        Tumor   \n",
       "1                 EA_Designation = 86          EA_Designation = 87   \n",
       "2                        %_Tumor = 50                 %_Tumor = 90   \n",
       "3                        Race = White                 Race = White   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  Histo_Desc = Carcinoma, NOS   \n",
       "\n",
       "                            GSM190490                           GSM190491  \\\n",
       "0                               Tumor                               Tumor   \n",
       "1                 EA_Designation = 88                 EA_Designation = 89   \n",
       "2                        %_Tumor = 60                        %_Tumor = 60   \n",
       "3                        Race = Black                    Race = Am Indian   \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  Histo_Desc = Ductal Carcinoma, NOS   \n",
       "\n",
       "                            GSM190492  \n",
       "0                               Tumor  \n",
       "1                 EA_Designation = 90  \n",
       "2                        %_Tumor = 60  \n",
       "3                        Race = White  \n",
       "4  Histo_Desc = Ductal Carcinoma, NOS  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', '%_Tumor = 80', '%_Tumor = 60',\n",
       "       '%_Tumor = 70', '%_Tumor = 50', '%_Tumor = 90'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[2]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    # Check if the trait indicates a normal disease status\n",
    "    if (trait == 'diabetes reversal status: Yes'):\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(':')[1])/12\n",
    "        age = round(age, 0)\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM190415</th>\n",
       "      <th>GSM190416</th>\n",
       "      <th>GSM190417</th>\n",
       "      <th>GSM190418</th>\n",
       "      <th>GSM190419</th>\n",
       "      <th>GSM190420</th>\n",
       "      <th>GSM190421</th>\n",
       "      <th>GSM190422</th>\n",
       "      <th>GSM190423</th>\n",
       "      <th>GSM190424</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM190483</th>\n",
       "      <th>GSM190484</th>\n",
       "      <th>GSM190485</th>\n",
       "      <th>GSM190486</th>\n",
       "      <th>GSM190487</th>\n",
       "      <th>GSM190488</th>\n",
       "      <th>GSM190489</th>\n",
       "      <th>GSM190490</th>\n",
       "      <th>GSM190491</th>\n",
       "      <th>GSM190492</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Paget's Disease of bone</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         GSM190415  GSM190416  GSM190417  GSM190418  \\\n",
       "Paget's Disease of bone          1          1          1          1   \n",
       "\n",
       "                         GSM190419  GSM190420  GSM190421  GSM190422  \\\n",
       "Paget's Disease of bone          1          1          1          1   \n",
       "\n",
       "                         GSM190423  GSM190424  ...  GSM190483  GSM190484  \\\n",
       "Paget's Disease of bone          1          1  ...          1          1   \n",
       "\n",
       "                         GSM190485  GSM190486  GSM190487  GSM190488  \\\n",
       "Paget's Disease of bone          1          1          1          1   \n",
       "\n",
       "                         GSM190489  GSM190490  GSM190491  GSM190492  \n",
       "Paget's Disease of bone          1          1          1          1  \n",
       "\n",
       "[1 rows x 78 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM190415</th>\n",
       "      <th>GSM190416</th>\n",
       "      <th>GSM190417</th>\n",
       "      <th>GSM190418</th>\n",
       "      <th>GSM190419</th>\n",
       "      <th>GSM190420</th>\n",
       "      <th>GSM190421</th>\n",
       "      <th>GSM190422</th>\n",
       "      <th>GSM190423</th>\n",
       "      <th>GSM190424</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM190483</th>\n",
       "      <th>GSM190484</th>\n",
       "      <th>GSM190485</th>\n",
       "      <th>GSM190486</th>\n",
       "      <th>GSM190487</th>\n",
       "      <th>GSM190488</th>\n",
       "      <th>GSM190489</th>\n",
       "      <th>GSM190490</th>\n",
       "      <th>GSM190491</th>\n",
       "      <th>GSM190492</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1000_at</th>\n",
       "      <td>1469.4</td>\n",
       "      <td>1213.7</td>\n",
       "      <td>1926.6</td>\n",
       "      <td>932.1</td>\n",
       "      <td>1314.2</td>\n",
       "      <td>2640.2</td>\n",
       "      <td>796.4</td>\n",
       "      <td>1562.6</td>\n",
       "      <td>1501.3</td>\n",
       "      <td>1157.8</td>\n",
       "      <td>...</td>\n",
       "      <td>1013.7</td>\n",
       "      <td>947.7</td>\n",
       "      <td>1393.4</td>\n",
       "      <td>1444.9</td>\n",
       "      <td>1289.9</td>\n",
       "      <td>1268.8</td>\n",
       "      <td>1068.0</td>\n",
       "      <td>1515.5</td>\n",
       "      <td>1348.4</td>\n",
       "      <td>1283.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001_at</th>\n",
       "      <td>201.7</td>\n",
       "      <td>109.4</td>\n",
       "      <td>117.1</td>\n",
       "      <td>312.8</td>\n",
       "      <td>111.8</td>\n",
       "      <td>147.2</td>\n",
       "      <td>117.5</td>\n",
       "      <td>540.2</td>\n",
       "      <td>105.6</td>\n",
       "      <td>184.5</td>\n",
       "      <td>...</td>\n",
       "      <td>115.3</td>\n",
       "      <td>167.5</td>\n",
       "      <td>276.5</td>\n",
       "      <td>143.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>315.4</td>\n",
       "      <td>324.0</td>\n",
       "      <td>249.1</td>\n",
       "      <td>348.9</td>\n",
       "      <td>265.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1002_f_at</th>\n",
       "      <td>7.2</td>\n",
       "      <td>21.7</td>\n",
       "      <td>24.8</td>\n",
       "      <td>26.9</td>\n",
       "      <td>26.7</td>\n",
       "      <td>17.6</td>\n",
       "      <td>11.6</td>\n",
       "      <td>80.6</td>\n",
       "      <td>22.4</td>\n",
       "      <td>29.1</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>21.4</td>\n",
       "      <td>24.1</td>\n",
       "      <td>17.8</td>\n",
       "      <td>36.4</td>\n",
       "      <td>16.2</td>\n",
       "      <td>19.7</td>\n",
       "      <td>17.9</td>\n",
       "      <td>21.8</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1003_s_at</th>\n",
       "      <td>209.1</td>\n",
       "      <td>161.0</td>\n",
       "      <td>153.1</td>\n",
       "      <td>528.8</td>\n",
       "      <td>94.8</td>\n",
       "      <td>72.8</td>\n",
       "      <td>46.0</td>\n",
       "      <td>334.9</td>\n",
       "      <td>212.6</td>\n",
       "      <td>221.4</td>\n",
       "      <td>...</td>\n",
       "      <td>253.8</td>\n",
       "      <td>57.8</td>\n",
       "      <td>44.6</td>\n",
       "      <td>259.3</td>\n",
       "      <td>154.1</td>\n",
       "      <td>70.2</td>\n",
       "      <td>295.8</td>\n",
       "      <td>41.9</td>\n",
       "      <td>197.7</td>\n",
       "      <td>47.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1004_at</th>\n",
       "      <td>139.6</td>\n",
       "      <td>379.7</td>\n",
       "      <td>173.7</td>\n",
       "      <td>755.9</td>\n",
       "      <td>171.7</td>\n",
       "      <td>190.7</td>\n",
       "      <td>160.8</td>\n",
       "      <td>367.3</td>\n",
       "      <td>167.2</td>\n",
       "      <td>139.7</td>\n",
       "      <td>...</td>\n",
       "      <td>224.6</td>\n",
       "      <td>239.0</td>\n",
       "      <td>204.6</td>\n",
       "      <td>766.5</td>\n",
       "      <td>1075.1</td>\n",
       "      <td>75.4</td>\n",
       "      <td>332.2</td>\n",
       "      <td>114.1</td>\n",
       "      <td>100.2</td>\n",
       "      <td>222.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM190415  GSM190416  GSM190417  GSM190418  GSM190419  GSM190420  \\\n",
       "ID                                                                            \n",
       "1000_at       1469.4     1213.7     1926.6      932.1     1314.2     2640.2   \n",
       "1001_at        201.7      109.4      117.1      312.8      111.8      147.2   \n",
       "1002_f_at        7.2       21.7       24.8       26.9       26.7       17.6   \n",
       "1003_s_at      209.1      161.0      153.1      528.8       94.8       72.8   \n",
       "1004_at        139.6      379.7      173.7      755.9      171.7      190.7   \n",
       "\n",
       "           GSM190421  GSM190422  GSM190423  GSM190424  ...  GSM190483  \\\n",
       "ID                                                     ...              \n",
       "1000_at        796.4     1562.6     1501.3     1157.8  ...     1013.7   \n",
       "1001_at        117.5      540.2      105.6      184.5  ...      115.3   \n",
       "1002_f_at       11.6       80.6       22.4       29.1  ...       24.0   \n",
       "1003_s_at       46.0      334.9      212.6      221.4  ...      253.8   \n",
       "1004_at        160.8      367.3      167.2      139.7  ...      224.6   \n",
       "\n",
       "           GSM190484  GSM190485  GSM190486  GSM190487  GSM190488  GSM190489  \\\n",
       "ID                                                                            \n",
       "1000_at        947.7     1393.4     1444.9     1289.9     1268.8     1068.0   \n",
       "1001_at        167.5      276.5      143.0       98.0      315.4      324.0   \n",
       "1002_f_at       21.4       24.1       17.8       36.4       16.2       19.7   \n",
       "1003_s_at       57.8       44.6      259.3      154.1       70.2      295.8   \n",
       "1004_at        239.0      204.6      766.5     1075.1       75.4      332.2   \n",
       "\n",
       "           GSM190490  GSM190491  GSM190492  \n",
       "ID                                          \n",
       "1000_at       1515.5     1348.4     1283.7  \n",
       "1001_at        249.1      348.9      265.3  \n",
       "1002_f_at       17.9       21.8       16.0  \n",
       "1003_s_at       41.9      197.7       47.9  \n",
       "1004_at        114.1      100.2      222.6  \n",
       "\n",
       "[5 rows x 78 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1000_at', '1001_at', '1002_f_at', '1003_s_at', '1004_at'], 'GB_ACC': ['X60188', 'X60957', 'X65962', 'X68149', 'X68149'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['GenBank', 'GenBank', 'GenBank', 'GenBank', 'GenBank'], 'Target Description': ['X60188 /FEATURE=mRNA /DEFINITION=HSERK1 Human ERK1 mRNA for protein serine/threonine kinase', 'X60957 /FEATURE=cds /DEFINITION=HSTIEMR Human tie mRNA for putative receptor tyrosine kinase', 'X65962 /FEATURE=cds /DEFINITION=HSCP450 H.sapiens mRNA for cytochrome P-450', \"X68149 /FEATURE=cds /DEFINITION=HSBLR1A Homo sapiens BLR1 gene for Burkitt's lymphoma receptor 1\", \"X68149 /FEATURE=exon#2 /DEFINITION=HSBLR1A Homo sapiens BLR1 gene for Burkitt's lymphoma receptor 1\"], 'Representative Public ID': ['X60188', 'X60957', 'X65962', 'X68149', 'X68149'], 'Gene Title': ['mitogen-activated protein kinase 3', 'tyrosine kinase with immunoglobulin-like and EGF-like domains 1', 'cytochrome P450, family 2, subfamily C, polypeptide 19', 'chemokine (C-X-C motif) receptor 5', 'chemokine (C-X-C motif) receptor 5'], 'Gene Symbol': ['MAPK3', 'TIE1', 'CYP2C19', 'CXCR5', 'CXCR5'], 'ENTREZ_GENE_ID': ['5595', '7075', '1557', '643', '643'], 'RefSeq Transcript ID': ['NM_001040056 /// NM_001109891 /// NM_002746 /// XR_243293', 'NM_001253357 /// NM_005424 /// XM_005271163 /// XM_005271164 /// XM_006710869', 'NM_000769', 'NM_001716 /// NM_032966', 'NM_001716 /// NM_032966'], 'Gene Ontology Biological Process': ['0000165 // MAPK cascade // non-traceable author statement /// 0000165 // MAPK cascade // traceable author statement /// 0000186 // activation of MAPKK activity // traceable author statement /// 0000187 // activation of MAPK activity // traceable author statement /// 0001934 // positive regulation of protein phosphorylation // inferred from mutant phenotype /// 0002224 // toll-like receptor signaling pathway // traceable author statement /// 0002755 // MyD88-dependent toll-like receptor signaling pathway // traceable author statement /// 0002756 // MyD88-independent toll-like receptor signaling pathway // traceable author statement /// 0006351 // transcription, DNA-templated // inferred from electronic annotation /// 0006360 // transcription from RNA polymerase I promoter // traceable author statement /// 0006361 // transcription initiation from RNA polymerase I promoter // traceable author statement /// 0006468 // protein phosphorylation // inferred from direct assay /// 0006915 // apoptotic process // inferred from electronic annotation /// 0006974 // cellular response to DNA damage stimulus // inferred from electronic annotation /// 0006975 // DNA damage induced protein phosphorylation // inferred from direct assay /// 0007049 // cell cycle // inferred from electronic annotation /// 0007173 // epidermal growth factor receptor signaling pathway // traceable author statement /// 0007264 // small GTPase mediated signal transduction // traceable author statement /// 0007265 // Ras protein signal transduction // traceable author statement /// 0007411 // axon guidance // traceable author statement /// 0007596 // blood coagulation // traceable author statement /// 0008286 // insulin receptor signaling pathway // traceable author statement /// 0008543 // fibroblast growth factor receptor signaling pathway // traceable author statement /// 0009887 // organ morphogenesis // inferred from electronic annotation /// 0010467 // gene expression // traceable author statement /// 0016032 // viral process // inferred from electronic annotation /// 0016310 // phosphorylation // inferred from direct assay /// 0019048 // modulation by virus of host morphology or physiology // inferred from electronic annotation /// 0019221 // cytokine-mediated signaling pathway // traceable author statement /// 0019233 // sensory perception of pain // inferred from electronic annotation /// 0030168 // platelet activation // traceable author statement /// 0030509 // BMP signaling pathway // inferred from mutant phenotype /// 0031663 // lipopolysaccharide-mediated signaling pathway // inferred from electronic annotation /// 0032496 // response to lipopolysaccharide // inferred from electronic annotation /// 0032872 // regulation of stress-activated MAPK cascade // traceable author statement /// 0033129 // positive regulation of histone phosphorylation // inferred from mutant phenotype /// 0034134 // toll-like receptor 2 signaling pathway // traceable author statement /// 0034138 // toll-like receptor 3 signaling pathway // traceable author statement /// 0034142 // toll-like receptor 4 signaling pathway // traceable author statement /// 0034146 // toll-like receptor 5 signaling pathway // traceable author statement /// 0034162 // toll-like receptor 9 signaling pathway // traceable author statement /// 0034166 // toll-like receptor 10 signaling pathway // traceable author statement /// 0035066 // positive regulation of histone acetylation // inferred from mutant phenotype /// 0035666 // TRIF-dependent toll-like receptor signaling pathway // traceable author statement /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0038095 // Fc-epsilon receptor signaling pathway // traceable author statement /// 0038096 // Fc-gamma receptor signaling pathway involved in phagocytosis // traceable author statement /// 0038123 // toll-like receptor TLR1:TLR2 signaling pathway // traceable author statement /// 0038124 // toll-like receptor TLR6:TLR2 signaling pathway // traceable author statement /// 0043330 // response to exogenous dsRNA // inferred from electronic annotation /// 0045087 // innate immune response // traceable author statement /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from mutant phenotype /// 0048011 // neurotrophin TRK receptor signaling pathway // traceable author statement /// 0051090 // regulation of sequence-specific DNA binding transcription factor activity // traceable author statement /// 0051216 // cartilage development // inferred from electronic annotation /// 0051403 // stress-activated MAPK cascade // traceable author statement /// 0051493 // regulation of cytoskeleton organization // traceable author statement /// 0060397 // JAK-STAT cascade involved in growth hormone signaling pathway // traceable author statement /// 0070374 // positive regulation of ERK1 and ERK2 cascade // inferred from mutant phenotype /// 0070498 // interleukin-1-mediated signaling pathway // inferred from mutant phenotype /// 0070849 // response to epidermal growth factor // inferred from direct assay /// 0071260 // cellular response to mechanical stimulus // inferred from expression pattern /// 0072584 // caveolin-mediated endocytosis // traceable author statement /// 0090170 // regulation of Golgi inheritance // traceable author statement /// 2000641 // regulation of early endosome to late endosome transport // traceable author statement /// 2000657 // negative regulation of apolipoprotein binding // inferred from electronic annotation', '0001525 // angiogenesis // inferred from electronic annotation /// 0001570 // vasculogenesis // inferred from electronic annotation /// 0001701 // in utero embryonic development // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007165 // signal transduction // traceable author statement /// 0007498 // mesoderm development // traceable author statement /// 0016310 // phosphorylation // inferred from electronic annotation /// 0016525 // negative regulation of angiogenesis // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // traceable author statement /// 0030336 // negative regulation of cell migration // inferred from electronic annotation /// 0032526 // response to retinoic acid // inferred from electronic annotation /// 0045026 // plasma membrane fusion // inferred from electronic annotation', '0006805 // xenobiotic metabolic process // traceable author statement /// 0008202 // steroid metabolic process // inferred from mutant phenotype /// 0016098 // monoterpenoid metabolic process // inferred from direct assay /// 0017144 // drug metabolic process // inferred from direct assay /// 0019369 // arachidonic acid metabolic process // traceable author statement /// 0019373 // epoxygenase P450 pathway // traceable author statement /// 0042738 // exogenous drug catabolic process // inferred from direct assay /// 0044281 // small molecule metabolic process // traceable author statement /// 0046483 // heterocycle metabolic process // inferred from direct assay /// 0055114 // oxidation-reduction process // inferred from direct assay /// 0097267 // omega-hydroxylase P450 pathway // traceable author statement', '0006928 // cellular component movement // traceable author statement /// 0006935 // chemotaxis // inferred from electronic annotation /// 0006955 // immune response // inferred from electronic annotation /// 0007165 // signal transduction // inferred from electronic annotation /// 0007186 // G-protein coupled receptor signaling pathway // inferred from electronic annotation /// 0032467 // positive regulation of cytokinesis // inferred from mutant phenotype /// 0042113 // B cell activation // inferred from electronic annotation /// 0048535 // lymph node development // inferred from electronic annotation /// 0070098 // chemokine-mediated signaling pathway // inferred from electronic annotation', '0006928 // cellular component movement // traceable author statement /// 0006935 // chemotaxis // inferred from electronic annotation /// 0006955 // immune response // inferred from electronic annotation /// 0007165 // signal transduction // inferred from electronic annotation /// 0007186 // G-protein coupled receptor signaling pathway // inferred from electronic annotation /// 0032467 // positive regulation of cytokinesis // inferred from mutant phenotype /// 0042113 // B cell activation // inferred from electronic annotation /// 0048535 // lymph node development // inferred from electronic annotation /// 0070098 // chemokine-mediated signaling pathway // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005634 // nucleus // inferred from direct assay /// 0005634 // nucleus // traceable author statement /// 0005654 // nucleoplasm // traceable author statement /// 0005730 // nucleolus // inferred from direct assay /// 0005737 // cytoplasm // inferred from electronic annotation /// 0005739 // mitochondrion // traceable author statement /// 0005769 // early endosome // traceable author statement /// 0005770 // late endosome // traceable author statement /// 0005794 // Golgi apparatus // traceable author statement /// 0005829 // cytosol // traceable author statement /// 0005856 // cytoskeleton // traceable author statement /// 0005901 // caveola // traceable author statement /// 0005925 // focal adhesion // traceable author statement /// 0015630 // microtubule cytoskeleton // inferred from direct assay /// 0031143 // pseudopodium // inferred from electronic annotation /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005886 // plasma membrane // inferred from electronic annotation /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation', '0005783 // endoplasmic reticulum // inferred from electronic annotation /// 0005789 // endoplasmic reticulum membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0031090 // organelle membrane // inferred from electronic annotation /// 0043231 // intracellular membrane-bounded organelle // traceable author statement', '0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation', '0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0001784 // phosphotyrosine binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004674 // protein serine/threonine kinase activity // inferred from electronic annotation /// 0004707 // MAP kinase activity // inferred from direct assay /// 0004707 // MAP kinase activity // non-traceable author statement /// 0004707 // MAP kinase activity // traceable author statement /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0019902 // phosphatase binding // inferred from physical interaction', '0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation', '0004497 // monooxygenase activity // inferred from direct assay /// 0005506 // iron ion binding // inferred from electronic annotation /// 0008395 // steroid hydroxylase activity // inferred from mutant phenotype /// 0009055 // electron carrier activity // inferred from electronic annotation /// 0016491 // oxidoreductase activity // inferred from direct assay /// 0016705 // oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen // inferred from electronic annotation /// 0018675 // (S)-limonene 6-monooxygenase activity // inferred from electronic annotation /// 0018676 // (S)-limonene 7-monooxygenase activity // inferred from electronic annotation /// 0019825 // oxygen binding // traceable author statement /// 0019899 // enzyme binding // inferred from physical interaction /// 0020037 // heme binding // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation /// 0052741 // (R)-limonene 6-monooxygenase activity // inferred from electronic annotation', '0004871 // signal transducer activity // inferred from electronic annotation /// 0004930 // G-protein coupled receptor activity // traceable author statement /// 0004950 // chemokine receptor activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0016494 // C-X-C chemokine receptor activity // inferred from electronic annotation', '0004871 // signal transducer activity // inferred from electronic annotation /// 0004930 // G-protein coupled receptor activity // traceable author statement /// 0004950 // chemokine receptor activity // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0016494 // C-X-C chemokine receptor activity // inferred from electronic annotation']}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM190415</th>\n",
       "      <th>GSM190416</th>\n",
       "      <th>GSM190417</th>\n",
       "      <th>GSM190418</th>\n",
       "      <th>GSM190419</th>\n",
       "      <th>GSM190420</th>\n",
       "      <th>GSM190421</th>\n",
       "      <th>GSM190422</th>\n",
       "      <th>GSM190423</th>\n",
       "      <th>GSM190424</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM190483</th>\n",
       "      <th>GSM190484</th>\n",
       "      <th>GSM190485</th>\n",
       "      <th>GSM190486</th>\n",
       "      <th>GSM190487</th>\n",
       "      <th>GSM190488</th>\n",
       "      <th>GSM190489</th>\n",
       "      <th>GSM190490</th>\n",
       "      <th>GSM190491</th>\n",
       "      <th>GSM190492</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>45.00</td>\n",
       "      <td>21.20</td>\n",
       "      <td>27.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>28.40</td>\n",
       "      <td>9.00</td>\n",
       "      <td>114.70</td>\n",
       "      <td>231.4</td>\n",
       "      <td>14.50</td>\n",
       "      <td>34.30</td>\n",
       "      <td>...</td>\n",
       "      <td>35.30</td>\n",
       "      <td>18.10</td>\n",
       "      <td>9.10</td>\n",
       "      <td>14.40</td>\n",
       "      <td>167.8</td>\n",
       "      <td>26.20</td>\n",
       "      <td>22.20</td>\n",
       "      <td>12.30</td>\n",
       "      <td>50.2</td>\n",
       "      <td>53.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>60.56</td>\n",
       "      <td>112.84</td>\n",
       "      <td>96.5</td>\n",
       "      <td>149.2</td>\n",
       "      <td>77.16</td>\n",
       "      <td>85.84</td>\n",
       "      <td>78.98</td>\n",
       "      <td>214.6</td>\n",
       "      <td>78.68</td>\n",
       "      <td>99.08</td>\n",
       "      <td>...</td>\n",
       "      <td>91.36</td>\n",
       "      <td>103.46</td>\n",
       "      <td>84.64</td>\n",
       "      <td>90.62</td>\n",
       "      <td>125.2</td>\n",
       "      <td>114.26</td>\n",
       "      <td>138.88</td>\n",
       "      <td>120.46</td>\n",
       "      <td>126.2</td>\n",
       "      <td>91.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>937.60</td>\n",
       "      <td>685.40</td>\n",
       "      <td>851.6</td>\n",
       "      <td>544.0</td>\n",
       "      <td>658.30</td>\n",
       "      <td>751.10</td>\n",
       "      <td>557.80</td>\n",
       "      <td>845.6</td>\n",
       "      <td>665.80</td>\n",
       "      <td>812.50</td>\n",
       "      <td>...</td>\n",
       "      <td>661.20</td>\n",
       "      <td>1031.70</td>\n",
       "      <td>788.40</td>\n",
       "      <td>678.50</td>\n",
       "      <td>98.4</td>\n",
       "      <td>1058.50</td>\n",
       "      <td>1415.20</td>\n",
       "      <td>1308.70</td>\n",
       "      <td>930.6</td>\n",
       "      <td>1476.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AANAT</th>\n",
       "      <td>1149.40</td>\n",
       "      <td>1015.20</td>\n",
       "      <td>1053.1</td>\n",
       "      <td>1223.1</td>\n",
       "      <td>897.70</td>\n",
       "      <td>983.00</td>\n",
       "      <td>637.40</td>\n",
       "      <td>1951.7</td>\n",
       "      <td>942.10</td>\n",
       "      <td>842.90</td>\n",
       "      <td>...</td>\n",
       "      <td>1024.30</td>\n",
       "      <td>527.10</td>\n",
       "      <td>620.50</td>\n",
       "      <td>647.90</td>\n",
       "      <td>1241.6</td>\n",
       "      <td>748.60</td>\n",
       "      <td>1032.70</td>\n",
       "      <td>845.00</td>\n",
       "      <td>656.6</td>\n",
       "      <td>606.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AASDHPPT</th>\n",
       "      <td>344.20</td>\n",
       "      <td>282.70</td>\n",
       "      <td>553.7</td>\n",
       "      <td>269.4</td>\n",
       "      <td>477.30</td>\n",
       "      <td>516.30</td>\n",
       "      <td>634.40</td>\n",
       "      <td>215.2</td>\n",
       "      <td>424.60</td>\n",
       "      <td>487.30</td>\n",
       "      <td>...</td>\n",
       "      <td>626.70</td>\n",
       "      <td>573.10</td>\n",
       "      <td>661.70</td>\n",
       "      <td>603.30</td>\n",
       "      <td>500.5</td>\n",
       "      <td>437.90</td>\n",
       "      <td>507.70</td>\n",
       "      <td>728.20</td>\n",
       "      <td>369.1</td>\n",
       "      <td>752.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZW10</th>\n",
       "      <td>55.80</td>\n",
       "      <td>9.10</td>\n",
       "      <td>48.2</td>\n",
       "      <td>17.7</td>\n",
       "      <td>64.30</td>\n",
       "      <td>38.00</td>\n",
       "      <td>35.70</td>\n",
       "      <td>70.0</td>\n",
       "      <td>11.00</td>\n",
       "      <td>18.40</td>\n",
       "      <td>...</td>\n",
       "      <td>43.90</td>\n",
       "      <td>55.10</td>\n",
       "      <td>53.90</td>\n",
       "      <td>56.60</td>\n",
       "      <td>31.4</td>\n",
       "      <td>51.30</td>\n",
       "      <td>31.10</td>\n",
       "      <td>9.60</td>\n",
       "      <td>57.9</td>\n",
       "      <td>48.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWINT</th>\n",
       "      <td>837.90</td>\n",
       "      <td>1452.70</td>\n",
       "      <td>1270.4</td>\n",
       "      <td>287.0</td>\n",
       "      <td>2092.80</td>\n",
       "      <td>459.90</td>\n",
       "      <td>613.30</td>\n",
       "      <td>178.1</td>\n",
       "      <td>3487.60</td>\n",
       "      <td>1434.20</td>\n",
       "      <td>...</td>\n",
       "      <td>934.10</td>\n",
       "      <td>789.00</td>\n",
       "      <td>799.80</td>\n",
       "      <td>1127.30</td>\n",
       "      <td>809.8</td>\n",
       "      <td>365.40</td>\n",
       "      <td>1062.40</td>\n",
       "      <td>1714.00</td>\n",
       "      <td>533.2</td>\n",
       "      <td>1344.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>2342.90</td>\n",
       "      <td>1803.00</td>\n",
       "      <td>2707.1</td>\n",
       "      <td>3035.6</td>\n",
       "      <td>1779.30</td>\n",
       "      <td>1813.10</td>\n",
       "      <td>2838.50</td>\n",
       "      <td>1544.8</td>\n",
       "      <td>2684.30</td>\n",
       "      <td>1580.10</td>\n",
       "      <td>...</td>\n",
       "      <td>2023.80</td>\n",
       "      <td>3519.60</td>\n",
       "      <td>2961.00</td>\n",
       "      <td>1501.50</td>\n",
       "      <td>2178.4</td>\n",
       "      <td>1833.00</td>\n",
       "      <td>1372.70</td>\n",
       "      <td>3333.10</td>\n",
       "      <td>1846.1</td>\n",
       "      <td>3228.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>215.00</td>\n",
       "      <td>254.10</td>\n",
       "      <td>317.1</td>\n",
       "      <td>397.8</td>\n",
       "      <td>258.10</td>\n",
       "      <td>419.70</td>\n",
       "      <td>303.00</td>\n",
       "      <td>351.7</td>\n",
       "      <td>273.10</td>\n",
       "      <td>351.80</td>\n",
       "      <td>...</td>\n",
       "      <td>291.50</td>\n",
       "      <td>381.00</td>\n",
       "      <td>299.90</td>\n",
       "      <td>232.00</td>\n",
       "      <td>171.8</td>\n",
       "      <td>283.70</td>\n",
       "      <td>265.60</td>\n",
       "      <td>486.30</td>\n",
       "      <td>418.8</td>\n",
       "      <td>193.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>419.10</td>\n",
       "      <td>281.40</td>\n",
       "      <td>253.9</td>\n",
       "      <td>366.7</td>\n",
       "      <td>387.20</td>\n",
       "      <td>167.50</td>\n",
       "      <td>444.80</td>\n",
       "      <td>282.8</td>\n",
       "      <td>277.00</td>\n",
       "      <td>292.20</td>\n",
       "      <td>...</td>\n",
       "      <td>261.50</td>\n",
       "      <td>349.60</td>\n",
       "      <td>320.00</td>\n",
       "      <td>431.70</td>\n",
       "      <td>194.8</td>\n",
       "      <td>271.10</td>\n",
       "      <td>296.50</td>\n",
       "      <td>131.20</td>\n",
       "      <td>190.5</td>\n",
       "      <td>422.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8221 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM190415  GSM190416  GSM190417  GSM190418  GSM190419  GSM190420  \\\n",
       "AADAC         45.00      21.20       27.0       49.0      28.40       9.00   \n",
       "AAK1          60.56     112.84       96.5      149.2      77.16      85.84   \n",
       "AAMP         937.60     685.40      851.6      544.0     658.30     751.10   \n",
       "AANAT       1149.40    1015.20     1053.1     1223.1     897.70     983.00   \n",
       "AASDHPPT     344.20     282.70      553.7      269.4     477.30     516.30   \n",
       "...             ...        ...        ...        ...        ...        ...   \n",
       "ZW10          55.80       9.10       48.2       17.7      64.30      38.00   \n",
       "ZWINT        837.90    1452.70     1270.4      287.0    2092.80     459.90   \n",
       "ZYX         2342.90    1803.00     2707.1     3035.6    1779.30    1813.10   \n",
       "ZZEF1        215.00     254.10      317.1      397.8     258.10     419.70   \n",
       "ZZZ3         419.10     281.40      253.9      366.7     387.20     167.50   \n",
       "\n",
       "          GSM190421  GSM190422  GSM190423  GSM190424  ...  GSM190483  \\\n",
       "AADAC        114.70      231.4      14.50      34.30  ...      35.30   \n",
       "AAK1          78.98      214.6      78.68      99.08  ...      91.36   \n",
       "AAMP         557.80      845.6     665.80     812.50  ...     661.20   \n",
       "AANAT        637.40     1951.7     942.10     842.90  ...    1024.30   \n",
       "AASDHPPT     634.40      215.2     424.60     487.30  ...     626.70   \n",
       "...             ...        ...        ...        ...  ...        ...   \n",
       "ZW10          35.70       70.0      11.00      18.40  ...      43.90   \n",
       "ZWINT        613.30      178.1    3487.60    1434.20  ...     934.10   \n",
       "ZYX         2838.50     1544.8    2684.30    1580.10  ...    2023.80   \n",
       "ZZEF1        303.00      351.7     273.10     351.80  ...     291.50   \n",
       "ZZZ3         444.80      282.8     277.00     292.20  ...     261.50   \n",
       "\n",
       "          GSM190484  GSM190485  GSM190486  GSM190487  GSM190488  GSM190489  \\\n",
       "AADAC         18.10       9.10      14.40      167.8      26.20      22.20   \n",
       "AAK1         103.46      84.64      90.62      125.2     114.26     138.88   \n",
       "AAMP        1031.70     788.40     678.50       98.4    1058.50    1415.20   \n",
       "AANAT        527.10     620.50     647.90     1241.6     748.60    1032.70   \n",
       "AASDHPPT     573.10     661.70     603.30      500.5     437.90     507.70   \n",
       "...             ...        ...        ...        ...        ...        ...   \n",
       "ZW10          55.10      53.90      56.60       31.4      51.30      31.10   \n",
       "ZWINT        789.00     799.80    1127.30      809.8     365.40    1062.40   \n",
       "ZYX         3519.60    2961.00    1501.50     2178.4    1833.00    1372.70   \n",
       "ZZEF1        381.00     299.90     232.00      171.8     283.70     265.60   \n",
       "ZZZ3         349.60     320.00     431.70      194.8     271.10     296.50   \n",
       "\n",
       "          GSM190490  GSM190491  GSM190492  \n",
       "AADAC         12.30       50.2      53.80  \n",
       "AAK1         120.46      126.2      91.68  \n",
       "AAMP        1308.70      930.6    1476.80  \n",
       "AANAT        845.00      656.6     606.90  \n",
       "AASDHPPT     728.20      369.1     752.00  \n",
       "...             ...        ...        ...  \n",
       "ZW10           9.60       57.9      48.90  \n",
       "ZWINT       1714.00      533.2    1344.60  \n",
       "ZYX         3333.10     1846.1    3228.40  \n",
       "ZZEF1        486.30      418.8     193.10  \n",
       "ZZZ3         131.20      190.5     422.80  \n",
       "\n",
       "[8221 rows x 78 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Paget's Disease of bone</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>AANAT</th>\n",
       "      <th>AASDHPPT</th>\n",
       "      <th>AASS</th>\n",
       "      <th>AATF</th>\n",
       "      <th>AATK</th>\n",
       "      <th>ABAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZRSR2</th>\n",
       "      <th>ZSCAN12</th>\n",
       "      <th>ZSCAN26</th>\n",
       "      <th>ZSCAN9</th>\n",
       "      <th>ZSWIM8</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM190415</th>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>60.56</td>\n",
       "      <td>937.6</td>\n",
       "      <td>1149.4</td>\n",
       "      <td>344.2</td>\n",
       "      <td>46.0</td>\n",
       "      <td>1405.1</td>\n",
       "      <td>678.0</td>\n",
       "      <td>24.100000</td>\n",
       "      <td>...</td>\n",
       "      <td>279.05</td>\n",
       "      <td>49.15</td>\n",
       "      <td>290.3</td>\n",
       "      <td>239.8</td>\n",
       "      <td>64.6</td>\n",
       "      <td>55.8</td>\n",
       "      <td>837.9</td>\n",
       "      <td>2342.9</td>\n",
       "      <td>215.0</td>\n",
       "      <td>419.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190416</th>\n",
       "      <td>1.0</td>\n",
       "      <td>21.2</td>\n",
       "      <td>112.84</td>\n",
       "      <td>685.4</td>\n",
       "      <td>1015.2</td>\n",
       "      <td>282.7</td>\n",
       "      <td>94.5</td>\n",
       "      <td>2213.7</td>\n",
       "      <td>509.5</td>\n",
       "      <td>85.433333</td>\n",
       "      <td>...</td>\n",
       "      <td>233.80</td>\n",
       "      <td>60.05</td>\n",
       "      <td>222.9</td>\n",
       "      <td>210.3</td>\n",
       "      <td>91.9</td>\n",
       "      <td>9.1</td>\n",
       "      <td>1452.7</td>\n",
       "      <td>1803.0</td>\n",
       "      <td>254.1</td>\n",
       "      <td>281.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190417</th>\n",
       "      <td>1.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>96.50</td>\n",
       "      <td>851.6</td>\n",
       "      <td>1053.1</td>\n",
       "      <td>553.7</td>\n",
       "      <td>10.7</td>\n",
       "      <td>1526.2</td>\n",
       "      <td>329.7</td>\n",
       "      <td>70.333333</td>\n",
       "      <td>...</td>\n",
       "      <td>230.70</td>\n",
       "      <td>22.00</td>\n",
       "      <td>237.6</td>\n",
       "      <td>238.1</td>\n",
       "      <td>92.2</td>\n",
       "      <td>48.2</td>\n",
       "      <td>1270.4</td>\n",
       "      <td>2707.1</td>\n",
       "      <td>317.1</td>\n",
       "      <td>253.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190418</th>\n",
       "      <td>1.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>149.20</td>\n",
       "      <td>544.0</td>\n",
       "      <td>1223.1</td>\n",
       "      <td>269.4</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1085.3</td>\n",
       "      <td>1286.7</td>\n",
       "      <td>103.600000</td>\n",
       "      <td>...</td>\n",
       "      <td>671.75</td>\n",
       "      <td>45.70</td>\n",
       "      <td>179.4</td>\n",
       "      <td>341.1</td>\n",
       "      <td>232.7</td>\n",
       "      <td>17.7</td>\n",
       "      <td>287.0</td>\n",
       "      <td>3035.6</td>\n",
       "      <td>397.8</td>\n",
       "      <td>366.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190419</th>\n",
       "      <td>1.0</td>\n",
       "      <td>28.4</td>\n",
       "      <td>77.16</td>\n",
       "      <td>658.3</td>\n",
       "      <td>897.7</td>\n",
       "      <td>477.3</td>\n",
       "      <td>16.2</td>\n",
       "      <td>1859.8</td>\n",
       "      <td>580.0</td>\n",
       "      <td>53.566667</td>\n",
       "      <td>...</td>\n",
       "      <td>182.40</td>\n",
       "      <td>12.80</td>\n",
       "      <td>217.8</td>\n",
       "      <td>237.6</td>\n",
       "      <td>65.0</td>\n",
       "      <td>64.3</td>\n",
       "      <td>2092.8</td>\n",
       "      <td>1779.3</td>\n",
       "      <td>258.1</td>\n",
       "      <td>387.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190488</th>\n",
       "      <td>1.0</td>\n",
       "      <td>26.2</td>\n",
       "      <td>114.26</td>\n",
       "      <td>1058.5</td>\n",
       "      <td>748.6</td>\n",
       "      <td>437.9</td>\n",
       "      <td>29.9</td>\n",
       "      <td>1391.8</td>\n",
       "      <td>881.3</td>\n",
       "      <td>61.766667</td>\n",
       "      <td>...</td>\n",
       "      <td>331.00</td>\n",
       "      <td>52.20</td>\n",
       "      <td>268.4</td>\n",
       "      <td>331.0</td>\n",
       "      <td>102.3</td>\n",
       "      <td>51.3</td>\n",
       "      <td>365.4</td>\n",
       "      <td>1833.0</td>\n",
       "      <td>283.7</td>\n",
       "      <td>271.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190489</th>\n",
       "      <td>1.0</td>\n",
       "      <td>22.2</td>\n",
       "      <td>138.88</td>\n",
       "      <td>1415.2</td>\n",
       "      <td>1032.7</td>\n",
       "      <td>507.7</td>\n",
       "      <td>104.1</td>\n",
       "      <td>1334.4</td>\n",
       "      <td>848.4</td>\n",
       "      <td>28.233333</td>\n",
       "      <td>...</td>\n",
       "      <td>312.90</td>\n",
       "      <td>47.40</td>\n",
       "      <td>200.8</td>\n",
       "      <td>488.8</td>\n",
       "      <td>132.5</td>\n",
       "      <td>31.1</td>\n",
       "      <td>1062.4</td>\n",
       "      <td>1372.7</td>\n",
       "      <td>265.6</td>\n",
       "      <td>296.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190490</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12.3</td>\n",
       "      <td>120.46</td>\n",
       "      <td>1308.7</td>\n",
       "      <td>845.0</td>\n",
       "      <td>728.2</td>\n",
       "      <td>72.4</td>\n",
       "      <td>773.2</td>\n",
       "      <td>419.2</td>\n",
       "      <td>105.733333</td>\n",
       "      <td>...</td>\n",
       "      <td>266.00</td>\n",
       "      <td>59.65</td>\n",
       "      <td>210.1</td>\n",
       "      <td>293.1</td>\n",
       "      <td>143.7</td>\n",
       "      <td>9.6</td>\n",
       "      <td>1714.0</td>\n",
       "      <td>3333.1</td>\n",
       "      <td>486.3</td>\n",
       "      <td>131.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190491</th>\n",
       "      <td>1.0</td>\n",
       "      <td>50.2</td>\n",
       "      <td>126.20</td>\n",
       "      <td>930.6</td>\n",
       "      <td>656.6</td>\n",
       "      <td>369.1</td>\n",
       "      <td>19.1</td>\n",
       "      <td>1121.5</td>\n",
       "      <td>380.9</td>\n",
       "      <td>253.066667</td>\n",
       "      <td>...</td>\n",
       "      <td>284.45</td>\n",
       "      <td>46.60</td>\n",
       "      <td>323.2</td>\n",
       "      <td>550.7</td>\n",
       "      <td>80.2</td>\n",
       "      <td>57.9</td>\n",
       "      <td>533.2</td>\n",
       "      <td>1846.1</td>\n",
       "      <td>418.8</td>\n",
       "      <td>190.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM190492</th>\n",
       "      <td>1.0</td>\n",
       "      <td>53.8</td>\n",
       "      <td>91.68</td>\n",
       "      <td>1476.8</td>\n",
       "      <td>606.9</td>\n",
       "      <td>752.0</td>\n",
       "      <td>45.8</td>\n",
       "      <td>1265.9</td>\n",
       "      <td>563.3</td>\n",
       "      <td>75.600000</td>\n",
       "      <td>...</td>\n",
       "      <td>207.20</td>\n",
       "      <td>24.20</td>\n",
       "      <td>186.5</td>\n",
       "      <td>242.7</td>\n",
       "      <td>64.7</td>\n",
       "      <td>48.9</td>\n",
       "      <td>1344.6</td>\n",
       "      <td>3228.4</td>\n",
       "      <td>193.1</td>\n",
       "      <td>422.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 8222 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Paget's Disease of bone  AADAC    AAK1    AAMP   AANAT  AASDHPPT  \\\n",
       "GSM190415                      1.0   45.0   60.56   937.6  1149.4     344.2   \n",
       "GSM190416                      1.0   21.2  112.84   685.4  1015.2     282.7   \n",
       "GSM190417                      1.0   27.0   96.50   851.6  1053.1     553.7   \n",
       "GSM190418                      1.0   49.0  149.20   544.0  1223.1     269.4   \n",
       "GSM190419                      1.0   28.4   77.16   658.3   897.7     477.3   \n",
       "...                            ...    ...     ...     ...     ...       ...   \n",
       "GSM190488                      1.0   26.2  114.26  1058.5   748.6     437.9   \n",
       "GSM190489                      1.0   22.2  138.88  1415.2  1032.7     507.7   \n",
       "GSM190490                      1.0   12.3  120.46  1308.7   845.0     728.2   \n",
       "GSM190491                      1.0   50.2  126.20   930.6   656.6     369.1   \n",
       "GSM190492                      1.0   53.8   91.68  1476.8   606.9     752.0   \n",
       "\n",
       "            AASS    AATF    AATK        ABAT  ...   ZRSR2  ZSCAN12  ZSCAN26  \\\n",
       "GSM190415   46.0  1405.1   678.0   24.100000  ...  279.05    49.15    290.3   \n",
       "GSM190416   94.5  2213.7   509.5   85.433333  ...  233.80    60.05    222.9   \n",
       "GSM190417   10.7  1526.2   329.7   70.333333  ...  230.70    22.00    237.6   \n",
       "GSM190418   34.0  1085.3  1286.7  103.600000  ...  671.75    45.70    179.4   \n",
       "GSM190419   16.2  1859.8   580.0   53.566667  ...  182.40    12.80    217.8   \n",
       "...          ...     ...     ...         ...  ...     ...      ...      ...   \n",
       "GSM190488   29.9  1391.8   881.3   61.766667  ...  331.00    52.20    268.4   \n",
       "GSM190489  104.1  1334.4   848.4   28.233333  ...  312.90    47.40    200.8   \n",
       "GSM190490   72.4   773.2   419.2  105.733333  ...  266.00    59.65    210.1   \n",
       "GSM190491   19.1  1121.5   380.9  253.066667  ...  284.45    46.60    323.2   \n",
       "GSM190492   45.8  1265.9   563.3   75.600000  ...  207.20    24.20    186.5   \n",
       "\n",
       "           ZSCAN9  ZSWIM8  ZW10   ZWINT     ZYX  ZZEF1   ZZZ3  \n",
       "GSM190415   239.8    64.6  55.8   837.9  2342.9  215.0  419.1  \n",
       "GSM190416   210.3    91.9   9.1  1452.7  1803.0  254.1  281.4  \n",
       "GSM190417   238.1    92.2  48.2  1270.4  2707.1  317.1  253.9  \n",
       "GSM190418   341.1   232.7  17.7   287.0  3035.6  397.8  366.7  \n",
       "GSM190419   237.6    65.0  64.3  2092.8  1779.3  258.1  387.2  \n",
       "...           ...     ...   ...     ...     ...    ...    ...  \n",
       "GSM190488   331.0   102.3  51.3   365.4  1833.0  283.7  271.1  \n",
       "GSM190489   488.8   132.5  31.1  1062.4  1372.7  265.6  296.5  \n",
       "GSM190490   293.1   143.7   9.6  1714.0  3333.1  486.3  131.2  \n",
       "GSM190491   550.7    80.2  57.9   533.2  1846.1  418.8  190.5  \n",
       "GSM190492   242.7    64.7  48.9  1344.6  3228.4  193.1  422.8  \n",
       "\n",
       "[78 rows x 8222 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 78 samples.\n",
      "For the feature 'Paget's Disease of bone', the least common label is '1.0' with 78 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Paget's Disease of bone' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output2\\Jiayi\\Paget's-Disease-of-bone\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
