{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = \"Sickle Cell Anemia\"\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE11372',\n",
       " 'GSE11524',\n",
       " 'GSE117613',\n",
       " 'GSE16728',\n",
       " 'GSE17078',\n",
       " 'GSE1849',\n",
       " 'GSE35007',\n",
       " 'GSE41575',\n",
       " 'GSE46471',\n",
       " 'GSE53441',\n",
       " 'GSE84632',\n",
       " 'GSE84633',\n",
       " 'GSE84634',\n",
       " 'GSE84635',\n",
       " 'GSE9877']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Sickle-Cell-Anemia\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE11372\\\\GSE11372_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE11372\\\\GSE11372-GPL1426_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE11372\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE11524\\\\GSE11524_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE11524\\\\GSE11524_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE11524\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE117613\\\\GSE117613_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE117613\\\\GSE117613_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE117613\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE16728\\\\GSE16728_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE16728\\\\GSE16728_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE16728\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE17078\\\\GSE17078_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE17078\\\\GSE17078_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE17078\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE1849\\\\GSE1849_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE1849\\\\GSE1849-GPL96_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# \n",
    "cohort = accession_num = \"GSE1849\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Empty\n",
    "cohort = accession_num = \"GSE35007\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE41575\\\\GSE41575_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE41575\\\\GSE41575-GPL14550_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE41575\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE46471\\\\GSE46471_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE46471\\\\GSE46471_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE46471\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE53441\\\\GSE53441_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE53441\\\\GSE53441_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE53441\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84632\\\\GSE84632_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84632\\\\GSE84632_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Biased\n",
    "cohort = accession_num = \"GSE84632\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84633\\\\GSE84633_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84633\\\\GSE84633_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Biased\n",
    "cohort = accession_num = \"GSE84633\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84634\\\\GSE84634_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84634\\\\GSE84634_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Biased\n",
    "cohort = accession_num = \"GSE84634\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84635\\\\GSE84635_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE84635\\\\GSE84635-GPL16686_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Biased\n",
    "cohort = accession_num = \"GSE84635\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE9877\\\\GSE9877_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Sickle-Cell-Anemia\\\\GSE9877\\\\GSE9877_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE9877\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Genetic Endothelial Systems Biology of Sickle Stroke Risk\"\n",
      "!Series_summary\t\"Genetic differences in endothelial biology could underlie development of phenotypic heterogeneity amongst individuals afflicted with vascular diseases. We obtained BOEC (blood outgrowth endothelial cells) from 20 subjects with sickle cell anemia (age 4-19) shown to be either at-risk (n=11) or not-at-risk (n=9) for ischemic stroke due to, respectively, having or not having occlusive disease at the Circle of Willis (CoW).\"\n",
      "!Series_summary\t\"Gene expression profiling identified no significant single gene differences between the two groups, as expected. However, analysis of Biological Systems Scores, using gene sets that were pre-determined to survey each of nine biological systems, showed that only changes in inflammation signaling are characteristic of the at-risk subjects, as supported by multiple statistical approaches\"\n",
      "!Series_summary\t\"Keywords: disease status analysis\"\n",
      "!Series_overall_design\t\"We obtained BOEC (blood outgrowth endothelial cells) from 20 subjects with sickle cell anemia (age 4-19) shown to be either at-risk (n=11) or not-at-risk (n=9) for ischemic stroke due to, respectively, having or not having occlusive disease at the Circle of Willis (CoW). To allow power calculations to be done, we performed microarray analysis on BOEC from 27 normal subjects of diverse ages. Gene expression profilings were obtained by using Affymetrix U133A chips\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM249425</th>\n",
       "      <th>GSM249426</th>\n",
       "      <th>GSM249427</th>\n",
       "      <th>GSM249428</th>\n",
       "      <th>GSM249429</th>\n",
       "      <th>GSM249430</th>\n",
       "      <th>GSM249431</th>\n",
       "      <th>GSM249432</th>\n",
       "      <th>GSM249433</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM249462</th>\n",
       "      <th>GSM249463</th>\n",
       "      <th>GSM249464</th>\n",
       "      <th>GSM249465</th>\n",
       "      <th>GSM249466</th>\n",
       "      <th>GSM249467</th>\n",
       "      <th>GSM249468</th>\n",
       "      <th>GSM249469</th>\n",
       "      <th>GSM249470</th>\n",
       "      <th>GSM249471</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>cell type: blood outgrowth endothelial cell</td>\n",
       "      <td>...</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "      <td>At Risk: +</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>status: normal</td>\n",
       "      <td>...</td>\n",
       "      <td>Age: 10</td>\n",
       "      <td>Age: 4</td>\n",
       "      <td>Age: 13</td>\n",
       "      <td>Age: 9</td>\n",
       "      <td>Age: 19</td>\n",
       "      <td>Age: 18</td>\n",
       "      <td>Age: 13</td>\n",
       "      <td>Age: 12</td>\n",
       "      <td>Age: 16</td>\n",
       "      <td>Age: 18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>race: Caucasian</td>\n",
       "      <td>race: African American</td>\n",
       "      <td>...</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "      <td>Sex: F</td>\n",
       "      <td>Sex: M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 50</td>\n",
       "      <td>age: 60</td>\n",
       "      <td>age: 53</td>\n",
       "      <td>age: 40</td>\n",
       "      <td>age: 41</td>\n",
       "      <td>age: 33</td>\n",
       "      <td>age: 23</td>\n",
       "      <td>age: 46</td>\n",
       "      <td>age: 31</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>sex: M</td>\n",
       "      <td>sex: F</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                                    GSM249425  \\\n",
       "0  !Sample_characteristics_ch1  cell type: blood outgrowth endothelial cell   \n",
       "1  !Sample_characteristics_ch1                               status: normal   \n",
       "2  !Sample_characteristics_ch1                       race: African American   \n",
       "3  !Sample_characteristics_ch1                                      age: 50   \n",
       "4  !Sample_characteristics_ch1                                       sex: F   \n",
       "\n",
       "                                     GSM249426  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                       race: African American   \n",
       "3                                      age: 60   \n",
       "4                                       sex: M   \n",
       "\n",
       "                                     GSM249427  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                       race: African American   \n",
       "3                                      age: 53   \n",
       "4                                       sex: F   \n",
       "\n",
       "                                     GSM249428  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                       race: African American   \n",
       "3                                      age: 40   \n",
       "4                                       sex: F   \n",
       "\n",
       "                                     GSM249429  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                       race: African American   \n",
       "3                                      age: 41   \n",
       "4                                       sex: M   \n",
       "\n",
       "                                     GSM249430  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                       race: African American   \n",
       "3                                      age: 33   \n",
       "4                                       sex: M   \n",
       "\n",
       "                                     GSM249431  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                       race: African American   \n",
       "3                                      age: 23   \n",
       "4                                       sex: F   \n",
       "\n",
       "                                     GSM249432  \\\n",
       "0  cell type: blood outgrowth endothelial cell   \n",
       "1                               status: normal   \n",
       "2                              race: Caucasian   \n",
       "3                                      age: 46   \n",
       "4                                       sex: M   \n",
       "\n",
       "                                     GSM249433  ...   GSM249462   GSM249463  \\\n",
       "0  cell type: blood outgrowth endothelial cell  ...  At Risk: +  At Risk: +   \n",
       "1                               status: normal  ...     Age: 10      Age: 4   \n",
       "2                       race: African American  ...      Sex: F      Sex: F   \n",
       "3                                      age: 31  ...         NaN         NaN   \n",
       "4                                       sex: F  ...         NaN         NaN   \n",
       "\n",
       "    GSM249464   GSM249465   GSM249466   GSM249467   GSM249468   GSM249469  \\\n",
       "0  At Risk: +  At Risk: +  At Risk: +  At Risk: +  At Risk: +  At Risk: +   \n",
       "1     Age: 13      Age: 9     Age: 19     Age: 18     Age: 13     Age: 12   \n",
       "2      Sex: M      Sex: F      Sex: F      Sex: M      Sex: F      Sex: M   \n",
       "3         NaN         NaN         NaN         NaN         NaN         NaN   \n",
       "4         NaN         NaN         NaN         NaN         NaN         NaN   \n",
       "\n",
       "    GSM249470   GSM249471  \n",
       "0  At Risk: +  At Risk: +  \n",
       "1     Age: 16     Age: 18  \n",
       "2      Sex: F      Sex: M  \n",
       "3         NaN         NaN  \n",
       "4         NaN         NaN  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1', 'status: normal', 'Age: 11',\n",
       "       'Age: 15', 'Age: 8', 'Age: 14', 'Age: 17', 'Age: 12', 'Age: 10',\n",
       "       'Age: 4', 'Age: 13', 'Age: 9', 'Age: 19', 'Age: 18', 'Age: 16'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    if (trait == 'diagnosis: normal'):\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(':')[1])\n",
    "        age = round(age, 0)\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1293592</th>\n",
       "      <th>GSM1293593</th>\n",
       "      <th>GSM1293594</th>\n",
       "      <th>GSM1293595</th>\n",
       "      <th>GSM1293596</th>\n",
       "      <th>GSM1293597</th>\n",
       "      <th>GSM1293598</th>\n",
       "      <th>GSM1293599</th>\n",
       "      <th>GSM1293600</th>\n",
       "      <th>GSM1293601</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1293616</th>\n",
       "      <th>GSM1293617</th>\n",
       "      <th>GSM1293618</th>\n",
       "      <th>GSM1293619</th>\n",
       "      <th>GSM1293620</th>\n",
       "      <th>GSM1293621</th>\n",
       "      <th>GSM1293622</th>\n",
       "      <th>GSM1293623</th>\n",
       "      <th>GSM1293624</th>\n",
       "      <th>GSM1293625</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sickle Cell Anemia</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM1293592  GSM1293593  GSM1293594  GSM1293595  \\\n",
       "Sickle Cell Anemia           1           1           1           1   \n",
       "\n",
       "                    GSM1293596  GSM1293597  GSM1293598  GSM1293599  \\\n",
       "Sickle Cell Anemia           1           1           1           1   \n",
       "\n",
       "                    GSM1293600  GSM1293601  ...  GSM1293616  GSM1293617  \\\n",
       "Sickle Cell Anemia           1           1  ...           0           0   \n",
       "\n",
       "                    GSM1293618  GSM1293619  GSM1293620  GSM1293621  \\\n",
       "Sickle Cell Anemia           0           0           0           0   \n",
       "\n",
       "                    GSM1293622  GSM1293623  GSM1293624  GSM1293625  \n",
       "Sickle Cell Anemia           0           0           0           0  \n",
       "\n",
       "[1 rows x 34 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1293592</th>\n",
       "      <th>GSM1293593</th>\n",
       "      <th>GSM1293594</th>\n",
       "      <th>GSM1293595</th>\n",
       "      <th>GSM1293596</th>\n",
       "      <th>GSM1293597</th>\n",
       "      <th>GSM1293598</th>\n",
       "      <th>GSM1293599</th>\n",
       "      <th>GSM1293600</th>\n",
       "      <th>GSM1293601</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1293616</th>\n",
       "      <th>GSM1293617</th>\n",
       "      <th>GSM1293618</th>\n",
       "      <th>GSM1293619</th>\n",
       "      <th>GSM1293620</th>\n",
       "      <th>GSM1293621</th>\n",
       "      <th>GSM1293622</th>\n",
       "      <th>GSM1293623</th>\n",
       "      <th>GSM1293624</th>\n",
       "      <th>GSM1293625</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>6.26</td>\n",
       "      <td>6.62</td>\n",
       "      <td>6.36</td>\n",
       "      <td>6.44</td>\n",
       "      <td>6.67</td>\n",
       "      <td>6.48</td>\n",
       "      <td>6.36</td>\n",
       "      <td>6.82</td>\n",
       "      <td>6.35</td>\n",
       "      <td>6.44</td>\n",
       "      <td>...</td>\n",
       "      <td>6.27</td>\n",
       "      <td>6.11</td>\n",
       "      <td>6.47</td>\n",
       "      <td>6.28</td>\n",
       "      <td>6.47</td>\n",
       "      <td>6.35</td>\n",
       "      <td>6.59</td>\n",
       "      <td>6.35</td>\n",
       "      <td>6.50</td>\n",
       "      <td>5.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>7.30</td>\n",
       "      <td>6.70</td>\n",
       "      <td>6.89</td>\n",
       "      <td>7.34</td>\n",
       "      <td>7.26</td>\n",
       "      <td>7.32</td>\n",
       "      <td>7.22</td>\n",
       "      <td>7.24</td>\n",
       "      <td>7.42</td>\n",
       "      <td>7.34</td>\n",
       "      <td>...</td>\n",
       "      <td>7.36</td>\n",
       "      <td>7.42</td>\n",
       "      <td>7.24</td>\n",
       "      <td>7.28</td>\n",
       "      <td>7.30</td>\n",
       "      <td>7.21</td>\n",
       "      <td>7.34</td>\n",
       "      <td>7.28</td>\n",
       "      <td>7.39</td>\n",
       "      <td>7.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>7.59</td>\n",
       "      <td>8.37</td>\n",
       "      <td>7.97</td>\n",
       "      <td>8.18</td>\n",
       "      <td>7.85</td>\n",
       "      <td>8.71</td>\n",
       "      <td>8.54</td>\n",
       "      <td>8.19</td>\n",
       "      <td>7.66</td>\n",
       "      <td>7.88</td>\n",
       "      <td>...</td>\n",
       "      <td>7.63</td>\n",
       "      <td>7.79</td>\n",
       "      <td>7.31</td>\n",
       "      <td>7.68</td>\n",
       "      <td>8.07</td>\n",
       "      <td>7.74</td>\n",
       "      <td>7.50</td>\n",
       "      <td>8.80</td>\n",
       "      <td>7.57</td>\n",
       "      <td>7.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>7.50</td>\n",
       "      <td>7.50</td>\n",
       "      <td>7.40</td>\n",
       "      <td>7.38</td>\n",
       "      <td>7.85</td>\n",
       "      <td>7.43</td>\n",
       "      <td>7.58</td>\n",
       "      <td>7.66</td>\n",
       "      <td>7.33</td>\n",
       "      <td>7.41</td>\n",
       "      <td>...</td>\n",
       "      <td>7.23</td>\n",
       "      <td>7.29</td>\n",
       "      <td>7.29</td>\n",
       "      <td>7.42</td>\n",
       "      <td>7.49</td>\n",
       "      <td>7.25</td>\n",
       "      <td>7.34</td>\n",
       "      <td>7.31</td>\n",
       "      <td>7.33</td>\n",
       "      <td>6.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>2.89</td>\n",
       "      <td>2.90</td>\n",
       "      <td>2.80</td>\n",
       "      <td>2.74</td>\n",
       "      <td>2.88</td>\n",
       "      <td>2.99</td>\n",
       "      <td>2.77</td>\n",
       "      <td>2.60</td>\n",
       "      <td>2.79</td>\n",
       "      <td>2.76</td>\n",
       "      <td>...</td>\n",
       "      <td>2.94</td>\n",
       "      <td>2.79</td>\n",
       "      <td>2.75</td>\n",
       "      <td>2.78</td>\n",
       "      <td>2.94</td>\n",
       "      <td>2.74</td>\n",
       "      <td>2.75</td>\n",
       "      <td>2.85</td>\n",
       "      <td>2.86</td>\n",
       "      <td>3.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM1293592  GSM1293593  GSM1293594  GSM1293595  GSM1293596  \\\n",
       "ID                                                                      \n",
       "1007_s_at        6.26        6.62        6.36        6.44        6.67   \n",
       "1053_at          7.30        6.70        6.89        7.34        7.26   \n",
       "117_at           7.59        8.37        7.97        8.18        7.85   \n",
       "121_at           7.50        7.50        7.40        7.38        7.85   \n",
       "1255_g_at        2.89        2.90        2.80        2.74        2.88   \n",
       "\n",
       "           GSM1293597  GSM1293598  GSM1293599  GSM1293600  GSM1293601  ...  \\\n",
       "ID                                                                     ...   \n",
       "1007_s_at        6.48        6.36        6.82        6.35        6.44  ...   \n",
       "1053_at          7.32        7.22        7.24        7.42        7.34  ...   \n",
       "117_at           8.71        8.54        8.19        7.66        7.88  ...   \n",
       "121_at           7.43        7.58        7.66        7.33        7.41  ...   \n",
       "1255_g_at        2.99        2.77        2.60        2.79        2.76  ...   \n",
       "\n",
       "           GSM1293616  GSM1293617  GSM1293618  GSM1293619  GSM1293620  \\\n",
       "ID                                                                      \n",
       "1007_s_at        6.27        6.11        6.47        6.28        6.47   \n",
       "1053_at          7.36        7.42        7.24        7.28        7.30   \n",
       "117_at           7.63        7.79        7.31        7.68        8.07   \n",
       "121_at           7.23        7.29        7.29        7.42        7.49   \n",
       "1255_g_at        2.94        2.79        2.75        2.78        2.94   \n",
       "\n",
       "           GSM1293621  GSM1293622  GSM1293623  GSM1293624  GSM1293625  \n",
       "ID                                                                     \n",
       "1007_s_at        6.35        6.59        6.35        6.50        5.94  \n",
       "1053_at          7.21        7.34        7.28        7.39        7.04  \n",
       "117_at           7.74        7.50        8.80        7.57        7.43  \n",
       "121_at           7.25        7.34        7.31        7.33        6.83  \n",
       "1255_g_at        2.74        2.75        2.85        2.86        3.01  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1007_s_at', '1053_at', '117_at', '121_at', '1255_g_at'], 'GB_ACC': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'SPOT_ID': [nan, nan, nan, nan, nan], 'Species Scientific Name': ['Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens'], 'Annotation Date': ['Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014', 'Oct 6, 2014'], 'Sequence Type': ['Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence', 'Exemplar sequence'], 'Sequence Source': ['Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database', 'GenBank', 'Affymetrix Proprietary Database'], 'Target Description': ['U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Human receptor tyrosine kinase DDR gene, complete cds', 'M87338 /FEATURE= /DEFINITION=HUMA1SBU Human replication factor C, 40-kDa subunit (A1) mRNA, complete cds', \"X51757 /FEATURE=cds /DEFINITION=HSP70B Human heat-shock protein HSP70B' gene\", 'X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens Pax8 mRNA', 'L36861 /FEATURE=expanded_cds /DEFINITION=HUMGCAPB Homo sapiens guanylate cyclase activating protein (GCAP) gene exons 1-4, complete cds'], 'Representative Public ID': ['U48705', 'M87338', 'X51757', 'X69699', 'L36861'], 'Gene Title': ['discoidin domain receptor tyrosine kinase 1 /// microRNA 4640', 'replication factor C (activator 1) 2, 40kDa', \"heat shock 70kDa protein 6 (HSP70B')\", 'paired box 8', 'guanylate cyclase activator 1A (retina)'], 'Gene Symbol': ['DDR1 /// MIR4640', 'RFC2', 'HSPA6', 'PAX8', 'GUCA1A'], 'ENTREZ_GENE_ID': ['780 /// 100616237', '5982', '3310', '7849', '2978'], 'RefSeq Transcript ID': ['NM_001202521 /// NM_001202522 /// NM_001202523 /// NM_001954 /// NM_013993 /// NM_013994 /// NR_039783 /// XM_005249385 /// XM_005249386 /// XM_005249387 /// XM_005249389 /// XM_005272873 /// XM_005272874 /// XM_005272875 /// XM_005272877 /// XM_005275027 /// XM_005275028 /// XM_005275030 /// XM_005275031 /// XM_005275162 /// XM_005275163 /// XM_005275164 /// XM_005275166 /// XM_005275457 /// XM_005275458 /// XM_005275459 /// XM_005275461 /// XM_006715185 /// XM_006715186 /// XM_006715187 /// XM_006715188 /// XM_006715189 /// XM_006715190 /// XM_006725501 /// XM_006725502 /// XM_006725503 /// XM_006725504 /// XM_006725505 /// XM_006725506 /// XM_006725714 /// XM_006725715 /// XM_006725716 /// XM_006725717 /// XM_006725718 /// XM_006725719 /// XM_006725720 /// XM_006725721 /// XM_006725722 /// XM_006725827 /// XM_006725828 /// XM_006725829 /// XM_006725830 /// XM_006725831 /// XM_006725832 /// XM_006726017 /// XM_006726018 /// XM_006726019 /// XM_006726020 /// XM_006726021 /// XM_006726022 /// XR_427836 /// XR_430858 /// XR_430938 /// XR_430974 /// XR_431015', 'NM_001278791 /// NM_001278792 /// NM_001278793 /// NM_002914 /// NM_181471 /// XM_006716080', 'NM_002155', 'NM_003466 /// NM_013951 /// NM_013952 /// NM_013953 /// NM_013992', 'NM_000409 /// XM_006715073'], 'Gene Ontology Biological Process': ['0001558 // regulation of cell growth // inferred from electronic annotation /// 0001952 // regulation of cell-matrix adhesion // inferred from electronic annotation /// 0006468 // protein phosphorylation // inferred from electronic annotation /// 0007155 // cell adhesion // traceable author statement /// 0007169 // transmembrane receptor protein tyrosine kinase signaling pathway // inferred from electronic annotation /// 0007565 // female pregnancy // inferred from electronic annotation /// 0007566 // embryo implantation // inferred from electronic annotation /// 0007595 // lactation // inferred from electronic annotation /// 0008285 // negative regulation of cell proliferation // inferred from electronic annotation /// 0010715 // regulation of extracellular matrix disassembly // inferred from mutant phenotype /// 0014909 // smooth muscle cell migration // inferred from mutant phenotype /// 0016310 // phosphorylation // inferred from electronic annotation /// 0018108 // peptidyl-tyrosine phosphorylation // inferred from electronic annotation /// 0030198 // extracellular matrix organization // traceable author statement /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from direct assay /// 0038063 // collagen-activated tyrosine kinase receptor signaling pathway // inferred from mutant phenotype /// 0038083 // peptidyl-tyrosine autophosphorylation // inferred from direct assay /// 0043583 // ear development // inferred from electronic annotation /// 0044319 // wound healing, spreading of cells // inferred from mutant phenotype /// 0046777 // protein autophosphorylation // inferred from direct assay /// 0060444 // branching involved in mammary gland duct morphogenesis // inferred from electronic annotation /// 0060749 // mammary gland alveolus development // inferred from electronic annotation /// 0061302 // smooth muscle cell-matrix adhesion // inferred from mutant phenotype', '0000278 // mitotic cell cycle // traceable author statement /// 0000722 // telomere maintenance via recombination // traceable author statement /// 0000723 // telomere maintenance // traceable author statement /// 0006260 // DNA replication // traceable author statement /// 0006271 // DNA strand elongation involved in DNA replication // traceable author statement /// 0006281 // DNA repair // traceable author statement /// 0006283 // transcription-coupled nucleotide-excision repair // traceable author statement /// 0006289 // nucleotide-excision repair // traceable author statement /// 0006297 // nucleotide-excision repair, DNA gap filling // traceable author statement /// 0015979 // photosynthesis // inferred from electronic annotation /// 0015995 // chlorophyll biosynthetic process // inferred from electronic annotation /// 0032201 // telomere maintenance via semi-conservative replication // traceable author statement', '0000902 // cell morphogenesis // inferred from electronic annotation /// 0006200 // ATP catabolic process // inferred from direct assay /// 0006950 // response to stress // inferred from electronic annotation /// 0006986 // response to unfolded protein // traceable author statement /// 0034605 // cellular response to heat // inferred from direct assay /// 0042026 // protein refolding // inferred from direct assay /// 0070370 // cellular heat acclimation // inferred from mutant phenotype', '0001655 // urogenital system development // inferred from sequence or structural similarity /// 0001656 // metanephros development // inferred from electronic annotation /// 0001658 // branching involved in ureteric bud morphogenesis // inferred from expression pattern /// 0001822 // kidney development // inferred from expression pattern /// 0001823 // mesonephros development // inferred from sequence or structural similarity /// 0003337 // mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from expression pattern /// 0006351 // transcription, DNA-templated // inferred from direct assay /// 0006355 // regulation of transcription, DNA-templated // inferred from electronic annotation /// 0007275 // multicellular organismal development // inferred from electronic annotation /// 0007417 // central nervous system development // inferred from expression pattern /// 0009653 // anatomical structure morphogenesis // traceable author statement /// 0030154 // cell differentiation // inferred from electronic annotation /// 0030878 // thyroid gland development // inferred from expression pattern /// 0030878 // thyroid gland development // inferred from mutant phenotype /// 0038194 // thyroid-stimulating hormone signaling pathway // traceable author statement /// 0039003 // pronephric field specification // inferred from sequence or structural similarity /// 0042472 // inner ear morphogenesis // inferred from sequence or structural similarity /// 0042981 // regulation of apoptotic process // inferred from sequence or structural similarity /// 0045893 // positive regulation of transcription, DNA-templated // inferred from direct assay /// 0045893 // positive regulation of transcription, DNA-templated // inferred from sequence or structural similarity /// 0045944 // positive regulation of transcription from RNA polymerase II promoter // inferred from direct assay /// 0048793 // pronephros development // inferred from sequence or structural similarity /// 0071371 // cellular response to gonadotropin stimulus // inferred from direct assay /// 0071599 // otic vesicle development // inferred from expression pattern /// 0072050 // S-shaped body morphogenesis // inferred from electronic annotation /// 0072073 // kidney epithelium development // inferred from electronic annotation /// 0072108 // positive regulation of mesenchymal to epithelial transition involved in metanephros morphogenesis // inferred from sequence or structural similarity /// 0072164 // mesonephric tubule development // inferred from electronic annotation /// 0072207 // metanephric epithelium development // inferred from expression pattern /// 0072221 // metanephric distal convoluted tubule development // inferred from sequence or structural similarity /// 0072278 // metanephric comma-shaped body morphogenesis // inferred from expression pattern /// 0072284 // metanephric S-shaped body morphogenesis // inferred from expression pattern /// 0072289 // metanephric nephron tubule formation // inferred from sequence or structural similarity /// 0072305 // negative regulation of mesenchymal cell apoptotic process involved in metanephric nephron morphogenesis // inferred from sequence or structural similarity /// 0072307 // regulation of metanephric nephron tubule epithelial cell differentiation // inferred from sequence or structural similarity /// 0090190 // positive regulation of branching involved in ureteric bud morphogenesis // inferred from sequence or structural similarity /// 1900212 // negative regulation of mesenchymal cell apoptotic process involved in metanephros development // inferred from sequence or structural similarity /// 1900215 // negative regulation of apoptotic process involved in metanephric collecting duct development // inferred from sequence or structural similarity /// 1900218 // negative regulation of apoptotic process involved in metanephric nephron tubule development // inferred from sequence or structural similarity /// 2000594 // positive regulation of metanephric DCT cell differentiation // inferred from sequence or structural similarity /// 2000611 // positive regulation of thyroid hormone generation // inferred from mutant phenotype /// 2000612 // regulation of thyroid-stimulating hormone secretion // inferred from mutant phenotype', '0007165 // signal transduction // non-traceable author statement /// 0007601 // visual perception // inferred from electronic annotation /// 0007602 // phototransduction // inferred from electronic annotation /// 0007603 // phototransduction, visible light // traceable author statement /// 0016056 // rhodopsin mediated signaling pathway // traceable author statement /// 0022400 // regulation of rhodopsin mediated signaling pathway // traceable author statement /// 0030828 // positive regulation of cGMP biosynthetic process // inferred from electronic annotation /// 0031282 // regulation of guanylate cyclase activity // inferred from electronic annotation /// 0031284 // positive regulation of guanylate cyclase activity // inferred from electronic annotation /// 0050896 // response to stimulus // inferred from electronic annotation'], 'Gene Ontology Cellular Component': ['0005576 // extracellular region // inferred from electronic annotation /// 0005615 // extracellular space // inferred from direct assay /// 0005886 // plasma membrane // traceable author statement /// 0005887 // integral component of plasma membrane // traceable author statement /// 0016020 // membrane // inferred from electronic annotation /// 0016021 // integral component of membrane // inferred from electronic annotation /// 0043235 // receptor complex // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay', '0005634 // nucleus // inferred from electronic annotation /// 0005654 // nucleoplasm // traceable author statement /// 0005663 // DNA replication factor C complex // inferred from direct assay', '0005737 // cytoplasm // inferred from direct assay /// 0005814 // centriole // inferred from direct assay /// 0005829 // cytosol // inferred from direct assay /// 0008180 // COP9 signalosome // inferred from direct assay /// 0070062 // extracellular vesicular exosome // inferred from direct assay /// 0072562 // blood microparticle // inferred from direct assay', '0005634 // nucleus // inferred from direct assay /// 0005654 // nucleoplasm // inferred from sequence or structural similarity /// 0005730 // nucleolus // inferred from direct assay', '0001750 // photoreceptor outer segment // inferred from electronic annotation /// 0001917 // photoreceptor inner segment // inferred from electronic annotation /// 0005578 // proteinaceous extracellular matrix // inferred from electronic annotation /// 0005886 // plasma membrane // inferred from direct assay /// 0016020 // membrane // inferred from electronic annotation /// 0097381 // photoreceptor disc membrane // traceable author statement'], 'Gene Ontology Molecular Function': ['0000166 // nucleotide binding // inferred from electronic annotation /// 0004672 // protein kinase activity // inferred from electronic annotation /// 0004713 // protein tyrosine kinase activity // inferred from electronic annotation /// 0004714 // transmembrane receptor protein tyrosine kinase activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0005518 // collagen binding // inferred from direct assay /// 0005518 // collagen binding // inferred from mutant phenotype /// 0005524 // ATP binding // inferred from electronic annotation /// 0016301 // kinase activity // inferred from electronic annotation /// 0016740 // transferase activity // inferred from electronic annotation /// 0016772 // transferase activity, transferring phosphorus-containing groups // inferred from electronic annotation /// 0038062 // protein tyrosine kinase collagen receptor activity // inferred from direct assay /// 0046872 // metal ion binding // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0003677 // DNA binding // inferred from electronic annotation /// 0005515 // protein binding // inferred from physical interaction /// 0005524 // ATP binding // inferred from electronic annotation /// 0016851 // magnesium chelatase activity // inferred from electronic annotation /// 0017111 // nucleoside-triphosphatase activity // inferred from electronic annotation', '0000166 // nucleotide binding // inferred from electronic annotation /// 0005524 // ATP binding // inferred from electronic annotation /// 0019899 // enzyme binding // inferred from physical interaction /// 0031072 // heat shock protein binding // inferred from physical interaction /// 0042623 // ATPase activity, coupled // inferred from direct assay /// 0051082 // unfolded protein binding // inferred from direct assay', '0000979 // RNA polymerase II core promoter sequence-specific DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from direct assay /// 0003677 // DNA binding // inferred from mutant phenotype /// 0003700 // sequence-specific DNA binding transcription factor activity // inferred from direct assay /// 0004996 // thyroid-stimulating hormone receptor activity // traceable author statement /// 0005515 // protein binding // inferred from physical interaction /// 0044212 // transcription regulatory region DNA binding // inferred from direct assay', '0005509 // calcium ion binding // inferred from electronic annotation /// 0008048 // calcium sensitive guanylate cyclase activator activity // inferred from electronic annotation /// 0030249 // guanylate cyclase regulator activity // inferred from electronic annotation /// 0046872 // metal ion binding // inferred from electronic annotation']}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'GB_ACC', 'SPOT_ID', 'Species Scientific Name', 'Annotation Date',\n",
       "       'Sequence Type', 'Sequence Source', 'Target Description',\n",
       "       'Representative Public ID', 'Gene Title', 'Gene Symbol',\n",
       "       'ENTREZ_GENE_ID', 'RefSeq Transcript ID',\n",
       "       'Gene Ontology Biological Process', 'Gene Ontology Cellular Component',\n",
       "       'Gene Ontology Molecular Function'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Gene Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1293592</th>\n",
       "      <th>GSM1293593</th>\n",
       "      <th>GSM1293594</th>\n",
       "      <th>GSM1293595</th>\n",
       "      <th>GSM1293596</th>\n",
       "      <th>GSM1293597</th>\n",
       "      <th>GSM1293598</th>\n",
       "      <th>GSM1293599</th>\n",
       "      <th>GSM1293600</th>\n",
       "      <th>GSM1293601</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1293616</th>\n",
       "      <th>GSM1293617</th>\n",
       "      <th>GSM1293618</th>\n",
       "      <th>GSM1293619</th>\n",
       "      <th>GSM1293620</th>\n",
       "      <th>GSM1293621</th>\n",
       "      <th>GSM1293622</th>\n",
       "      <th>GSM1293623</th>\n",
       "      <th>GSM1293624</th>\n",
       "      <th>GSM1293625</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>6.790000</td>\n",
       "      <td>6.200000</td>\n",
       "      <td>6.580000</td>\n",
       "      <td>6.350000</td>\n",
       "      <td>6.240000</td>\n",
       "      <td>6.570000</td>\n",
       "      <td>6.530000</td>\n",
       "      <td>6.230000</td>\n",
       "      <td>6.630000</td>\n",
       "      <td>6.660000</td>\n",
       "      <td>...</td>\n",
       "      <td>6.570000</td>\n",
       "      <td>6.440000</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>7.180000</td>\n",
       "      <td>6.410000</td>\n",
       "      <td>6.650000</td>\n",
       "      <td>6.750000</td>\n",
       "      <td>6.430000</td>\n",
       "      <td>6.680000</td>\n",
       "      <td>6.270000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>4.730000</td>\n",
       "      <td>4.520000</td>\n",
       "      <td>4.650000</td>\n",
       "      <td>4.620000</td>\n",
       "      <td>4.210000</td>\n",
       "      <td>4.480000</td>\n",
       "      <td>4.780000</td>\n",
       "      <td>4.470000</td>\n",
       "      <td>4.530000</td>\n",
       "      <td>4.950000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.770000</td>\n",
       "      <td>4.780000</td>\n",
       "      <td>4.290000</td>\n",
       "      <td>4.870000</td>\n",
       "      <td>4.320000</td>\n",
       "      <td>4.240000</td>\n",
       "      <td>4.690000</td>\n",
       "      <td>4.630000</td>\n",
       "      <td>4.520000</td>\n",
       "      <td>4.790000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>4.110000</td>\n",
       "      <td>4.100000</td>\n",
       "      <td>4.215000</td>\n",
       "      <td>3.935000</td>\n",
       "      <td>4.035000</td>\n",
       "      <td>4.105000</td>\n",
       "      <td>4.245000</td>\n",
       "      <td>3.870000</td>\n",
       "      <td>4.115000</td>\n",
       "      <td>4.165000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.940000</td>\n",
       "      <td>3.835000</td>\n",
       "      <td>3.875000</td>\n",
       "      <td>3.860000</td>\n",
       "      <td>4.065000</td>\n",
       "      <td>4.080000</td>\n",
       "      <td>4.060000</td>\n",
       "      <td>4.045000</td>\n",
       "      <td>4.120000</td>\n",
       "      <td>3.860000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>4.005000</td>\n",
       "      <td>3.850000</td>\n",
       "      <td>3.970000</td>\n",
       "      <td>3.635000</td>\n",
       "      <td>4.090000</td>\n",
       "      <td>4.315000</td>\n",
       "      <td>3.905000</td>\n",
       "      <td>3.755000</td>\n",
       "      <td>3.835000</td>\n",
       "      <td>3.945000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.850000</td>\n",
       "      <td>3.725000</td>\n",
       "      <td>3.910000</td>\n",
       "      <td>3.705000</td>\n",
       "      <td>3.840000</td>\n",
       "      <td>3.970000</td>\n",
       "      <td>4.075000</td>\n",
       "      <td>3.790000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>4.010000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>5.370000</td>\n",
       "      <td>4.520000</td>\n",
       "      <td>5.530000</td>\n",
       "      <td>5.820000</td>\n",
       "      <td>4.990000</td>\n",
       "      <td>4.960000</td>\n",
       "      <td>5.020000</td>\n",
       "      <td>5.710000</td>\n",
       "      <td>5.720000</td>\n",
       "      <td>5.780000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.270000</td>\n",
       "      <td>5.830000</td>\n",
       "      <td>5.520000</td>\n",
       "      <td>4.380000</td>\n",
       "      <td>5.110000</td>\n",
       "      <td>5.420000</td>\n",
       "      <td>3.920000</td>\n",
       "      <td>6.420000</td>\n",
       "      <td>4.390000</td>\n",
       "      <td>5.650000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>4.320000</td>\n",
       "      <td>4.370000</td>\n",
       "      <td>4.230000</td>\n",
       "      <td>4.190000</td>\n",
       "      <td>4.280000</td>\n",
       "      <td>4.270000</td>\n",
       "      <td>4.310000</td>\n",
       "      <td>4.230000</td>\n",
       "      <td>4.270000</td>\n",
       "      <td>4.170000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.040000</td>\n",
       "      <td>3.990000</td>\n",
       "      <td>4.160000</td>\n",
       "      <td>4.410000</td>\n",
       "      <td>3.990000</td>\n",
       "      <td>4.180000</td>\n",
       "      <td>4.150000</td>\n",
       "      <td>3.830000</td>\n",
       "      <td>4.230000</td>\n",
       "      <td>4.310000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>7.133333</td>\n",
       "      <td>7.016667</td>\n",
       "      <td>6.723333</td>\n",
       "      <td>7.323333</td>\n",
       "      <td>6.870000</td>\n",
       "      <td>7.216667</td>\n",
       "      <td>7.013333</td>\n",
       "      <td>6.983333</td>\n",
       "      <td>7.260000</td>\n",
       "      <td>7.166667</td>\n",
       "      <td>...</td>\n",
       "      <td>7.260000</td>\n",
       "      <td>7.416667</td>\n",
       "      <td>7.513333</td>\n",
       "      <td>7.443333</td>\n",
       "      <td>7.456667</td>\n",
       "      <td>6.993333</td>\n",
       "      <td>7.333333</td>\n",
       "      <td>7.453333</td>\n",
       "      <td>7.236667</td>\n",
       "      <td>6.880000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>8.890000</td>\n",
       "      <td>9.260000</td>\n",
       "      <td>9.435000</td>\n",
       "      <td>9.610000</td>\n",
       "      <td>9.210000</td>\n",
       "      <td>9.730000</td>\n",
       "      <td>9.175000</td>\n",
       "      <td>9.150000</td>\n",
       "      <td>8.775000</td>\n",
       "      <td>9.240000</td>\n",
       "      <td>...</td>\n",
       "      <td>9.355000</td>\n",
       "      <td>9.370000</td>\n",
       "      <td>8.935000</td>\n",
       "      <td>9.160000</td>\n",
       "      <td>9.370000</td>\n",
       "      <td>9.335000</td>\n",
       "      <td>9.875000</td>\n",
       "      <td>9.465000</td>\n",
       "      <td>9.825000</td>\n",
       "      <td>9.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>5.856667</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>6.313333</td>\n",
       "      <td>6.313333</td>\n",
       "      <td>6.313333</td>\n",
       "      <td>6.206667</td>\n",
       "      <td>6.386667</td>\n",
       "      <td>6.343333</td>\n",
       "      <td>5.833333</td>\n",
       "      <td>6.220000</td>\n",
       "      <td>...</td>\n",
       "      <td>6.146667</td>\n",
       "      <td>6.200000</td>\n",
       "      <td>6.013333</td>\n",
       "      <td>6.076667</td>\n",
       "      <td>6.173333</td>\n",
       "      <td>6.320000</td>\n",
       "      <td>5.950000</td>\n",
       "      <td>6.276667</td>\n",
       "      <td>6.166667</td>\n",
       "      <td>6.403333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>7.495000</td>\n",
       "      <td>7.055000</td>\n",
       "      <td>7.050000</td>\n",
       "      <td>7.485000</td>\n",
       "      <td>6.985000</td>\n",
       "      <td>7.305000</td>\n",
       "      <td>7.110000</td>\n",
       "      <td>7.205000</td>\n",
       "      <td>7.585000</td>\n",
       "      <td>7.320000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.315000</td>\n",
       "      <td>7.410000</td>\n",
       "      <td>7.640000</td>\n",
       "      <td>7.595000</td>\n",
       "      <td>7.515000</td>\n",
       "      <td>7.225000</td>\n",
       "      <td>7.160000</td>\n",
       "      <td>7.365000</td>\n",
       "      <td>7.315000</td>\n",
       "      <td>7.230000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17857 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM1293592  GSM1293593  GSM1293594  GSM1293595  GSM1293596  \\\n",
       "A1BG        6.790000    6.200000    6.580000    6.350000    6.240000   \n",
       "A1BG-AS1    4.730000    4.520000    4.650000    4.620000    4.210000   \n",
       "A1CF        4.110000    4.100000    4.215000    3.935000    4.035000   \n",
       "A2M         4.005000    3.850000    3.970000    3.635000    4.090000   \n",
       "A2M-AS1     5.370000    4.520000    5.530000    5.820000    4.990000   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A      4.320000    4.370000    4.230000    4.190000    4.280000   \n",
       "ZYG11B      7.133333    7.016667    6.723333    7.323333    6.870000   \n",
       "ZYX         8.890000    9.260000    9.435000    9.610000    9.210000   \n",
       "ZZEF1       5.856667    6.500000    6.313333    6.313333    6.313333   \n",
       "ZZZ3        7.495000    7.055000    7.050000    7.485000    6.985000   \n",
       "\n",
       "          GSM1293597  GSM1293598  GSM1293599  GSM1293600  GSM1293601  ...  \\\n",
       "A1BG        6.570000    6.530000    6.230000    6.630000    6.660000  ...   \n",
       "A1BG-AS1    4.480000    4.780000    4.470000    4.530000    4.950000  ...   \n",
       "A1CF        4.105000    4.245000    3.870000    4.115000    4.165000  ...   \n",
       "A2M         4.315000    3.905000    3.755000    3.835000    3.945000  ...   \n",
       "A2M-AS1     4.960000    5.020000    5.710000    5.720000    5.780000  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "ZYG11A      4.270000    4.310000    4.230000    4.270000    4.170000  ...   \n",
       "ZYG11B      7.216667    7.013333    6.983333    7.260000    7.166667  ...   \n",
       "ZYX         9.730000    9.175000    9.150000    8.775000    9.240000  ...   \n",
       "ZZEF1       6.206667    6.386667    6.343333    5.833333    6.220000  ...   \n",
       "ZZZ3        7.305000    7.110000    7.205000    7.585000    7.320000  ...   \n",
       "\n",
       "          GSM1293616  GSM1293617  GSM1293618  GSM1293619  GSM1293620  \\\n",
       "A1BG        6.570000    6.440000    6.500000    7.180000    6.410000   \n",
       "A1BG-AS1    4.770000    4.780000    4.290000    4.870000    4.320000   \n",
       "A1CF        3.940000    3.835000    3.875000    3.860000    4.065000   \n",
       "A2M         3.850000    3.725000    3.910000    3.705000    3.840000   \n",
       "A2M-AS1     5.270000    5.830000    5.520000    4.380000    5.110000   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "ZYG11A      4.040000    3.990000    4.160000    4.410000    3.990000   \n",
       "ZYG11B      7.260000    7.416667    7.513333    7.443333    7.456667   \n",
       "ZYX         9.355000    9.370000    8.935000    9.160000    9.370000   \n",
       "ZZEF1       6.146667    6.200000    6.013333    6.076667    6.173333   \n",
       "ZZZ3        7.315000    7.410000    7.640000    7.595000    7.515000   \n",
       "\n",
       "          GSM1293621  GSM1293622  GSM1293623  GSM1293624  GSM1293625  \n",
       "A1BG        6.650000    6.750000    6.430000    6.680000    6.270000  \n",
       "A1BG-AS1    4.240000    4.690000    4.630000    4.520000    4.790000  \n",
       "A1CF        4.080000    4.060000    4.045000    4.120000    3.860000  \n",
       "A2M         3.970000    4.075000    3.790000    4.000000    4.010000  \n",
       "A2M-AS1     5.420000    3.920000    6.420000    4.390000    5.650000  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "ZYG11A      4.180000    4.150000    3.830000    4.230000    4.310000  \n",
       "ZYG11B      6.993333    7.333333    7.453333    7.236667    6.880000  \n",
       "ZYX         9.335000    9.875000    9.465000    9.825000    9.600000  \n",
       "ZZEF1       6.320000    5.950000    6.276667    6.166667    6.403333  \n",
       "ZZZ3        7.225000    7.160000    7.365000    7.315000    7.230000  \n",
       "\n",
       "[17857 rows x 34 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sickle Cell Anemia</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1293592</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.79</td>\n",
       "      <td>4.73</td>\n",
       "      <td>4.110</td>\n",
       "      <td>4.005</td>\n",
       "      <td>5.37</td>\n",
       "      <td>3.590</td>\n",
       "      <td>3.83</td>\n",
       "      <td>5.77</td>\n",
       "      <td>4.19</td>\n",
       "      <td>...</td>\n",
       "      <td>6.510</td>\n",
       "      <td>6.63</td>\n",
       "      <td>5.66</td>\n",
       "      <td>4.870</td>\n",
       "      <td>5.723333</td>\n",
       "      <td>4.32</td>\n",
       "      <td>7.133333</td>\n",
       "      <td>8.890</td>\n",
       "      <td>5.856667</td>\n",
       "      <td>7.495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293593</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.20</td>\n",
       "      <td>4.52</td>\n",
       "      <td>4.100</td>\n",
       "      <td>3.850</td>\n",
       "      <td>4.52</td>\n",
       "      <td>3.650</td>\n",
       "      <td>4.23</td>\n",
       "      <td>5.86</td>\n",
       "      <td>4.15</td>\n",
       "      <td>...</td>\n",
       "      <td>6.025</td>\n",
       "      <td>5.68</td>\n",
       "      <td>5.61</td>\n",
       "      <td>4.940</td>\n",
       "      <td>5.838333</td>\n",
       "      <td>4.37</td>\n",
       "      <td>7.016667</td>\n",
       "      <td>9.260</td>\n",
       "      <td>6.500000</td>\n",
       "      <td>7.055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293594</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.58</td>\n",
       "      <td>4.65</td>\n",
       "      <td>4.215</td>\n",
       "      <td>3.970</td>\n",
       "      <td>5.53</td>\n",
       "      <td>3.580</td>\n",
       "      <td>3.96</td>\n",
       "      <td>5.74</td>\n",
       "      <td>4.13</td>\n",
       "      <td>...</td>\n",
       "      <td>6.210</td>\n",
       "      <td>6.39</td>\n",
       "      <td>5.63</td>\n",
       "      <td>4.865</td>\n",
       "      <td>5.936667</td>\n",
       "      <td>4.23</td>\n",
       "      <td>6.723333</td>\n",
       "      <td>9.435</td>\n",
       "      <td>6.313333</td>\n",
       "      <td>7.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293595</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.35</td>\n",
       "      <td>4.62</td>\n",
       "      <td>3.935</td>\n",
       "      <td>3.635</td>\n",
       "      <td>5.82</td>\n",
       "      <td>3.590</td>\n",
       "      <td>3.94</td>\n",
       "      <td>5.48</td>\n",
       "      <td>4.00</td>\n",
       "      <td>...</td>\n",
       "      <td>6.370</td>\n",
       "      <td>6.41</td>\n",
       "      <td>6.10</td>\n",
       "      <td>4.900</td>\n",
       "      <td>6.178333</td>\n",
       "      <td>4.19</td>\n",
       "      <td>7.323333</td>\n",
       "      <td>9.610</td>\n",
       "      <td>6.313333</td>\n",
       "      <td>7.485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293596</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.24</td>\n",
       "      <td>4.21</td>\n",
       "      <td>4.035</td>\n",
       "      <td>4.090</td>\n",
       "      <td>4.99</td>\n",
       "      <td>3.800</td>\n",
       "      <td>4.11</td>\n",
       "      <td>5.92</td>\n",
       "      <td>3.90</td>\n",
       "      <td>...</td>\n",
       "      <td>6.135</td>\n",
       "      <td>6.52</td>\n",
       "      <td>5.60</td>\n",
       "      <td>4.790</td>\n",
       "      <td>6.088333</td>\n",
       "      <td>4.28</td>\n",
       "      <td>6.870000</td>\n",
       "      <td>9.210</td>\n",
       "      <td>6.313333</td>\n",
       "      <td>6.985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293597</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.57</td>\n",
       "      <td>4.48</td>\n",
       "      <td>4.105</td>\n",
       "      <td>4.315</td>\n",
       "      <td>4.96</td>\n",
       "      <td>3.620</td>\n",
       "      <td>3.77</td>\n",
       "      <td>5.67</td>\n",
       "      <td>4.18</td>\n",
       "      <td>...</td>\n",
       "      <td>6.460</td>\n",
       "      <td>6.50</td>\n",
       "      <td>5.45</td>\n",
       "      <td>4.885</td>\n",
       "      <td>5.725000</td>\n",
       "      <td>4.27</td>\n",
       "      <td>7.216667</td>\n",
       "      <td>9.730</td>\n",
       "      <td>6.206667</td>\n",
       "      <td>7.305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293598</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.53</td>\n",
       "      <td>4.78</td>\n",
       "      <td>4.245</td>\n",
       "      <td>3.905</td>\n",
       "      <td>5.02</td>\n",
       "      <td>3.445</td>\n",
       "      <td>4.38</td>\n",
       "      <td>5.72</td>\n",
       "      <td>3.91</td>\n",
       "      <td>...</td>\n",
       "      <td>5.245</td>\n",
       "      <td>5.98</td>\n",
       "      <td>5.40</td>\n",
       "      <td>4.785</td>\n",
       "      <td>6.140000</td>\n",
       "      <td>4.31</td>\n",
       "      <td>7.013333</td>\n",
       "      <td>9.175</td>\n",
       "      <td>6.386667</td>\n",
       "      <td>7.110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293599</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.23</td>\n",
       "      <td>4.47</td>\n",
       "      <td>3.870</td>\n",
       "      <td>3.755</td>\n",
       "      <td>5.71</td>\n",
       "      <td>3.465</td>\n",
       "      <td>4.29</td>\n",
       "      <td>5.31</td>\n",
       "      <td>3.71</td>\n",
       "      <td>...</td>\n",
       "      <td>6.180</td>\n",
       "      <td>6.11</td>\n",
       "      <td>5.87</td>\n",
       "      <td>4.900</td>\n",
       "      <td>6.186667</td>\n",
       "      <td>4.23</td>\n",
       "      <td>6.983333</td>\n",
       "      <td>9.150</td>\n",
       "      <td>6.343333</td>\n",
       "      <td>7.205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293600</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.63</td>\n",
       "      <td>4.53</td>\n",
       "      <td>4.115</td>\n",
       "      <td>3.835</td>\n",
       "      <td>5.72</td>\n",
       "      <td>3.315</td>\n",
       "      <td>3.89</td>\n",
       "      <td>5.68</td>\n",
       "      <td>4.39</td>\n",
       "      <td>...</td>\n",
       "      <td>6.640</td>\n",
       "      <td>6.78</td>\n",
       "      <td>6.04</td>\n",
       "      <td>5.000</td>\n",
       "      <td>5.970000</td>\n",
       "      <td>4.27</td>\n",
       "      <td>7.260000</td>\n",
       "      <td>8.775</td>\n",
       "      <td>5.833333</td>\n",
       "      <td>7.585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293601</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.66</td>\n",
       "      <td>4.95</td>\n",
       "      <td>4.165</td>\n",
       "      <td>3.945</td>\n",
       "      <td>5.78</td>\n",
       "      <td>3.525</td>\n",
       "      <td>3.99</td>\n",
       "      <td>5.92</td>\n",
       "      <td>4.01</td>\n",
       "      <td>...</td>\n",
       "      <td>6.455</td>\n",
       "      <td>6.15</td>\n",
       "      <td>5.49</td>\n",
       "      <td>4.860</td>\n",
       "      <td>5.916667</td>\n",
       "      <td>4.17</td>\n",
       "      <td>7.166667</td>\n",
       "      <td>9.240</td>\n",
       "      <td>6.220000</td>\n",
       "      <td>7.320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293602</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.63</td>\n",
       "      <td>4.65</td>\n",
       "      <td>3.840</td>\n",
       "      <td>4.220</td>\n",
       "      <td>5.35</td>\n",
       "      <td>3.565</td>\n",
       "      <td>4.75</td>\n",
       "      <td>5.68</td>\n",
       "      <td>3.95</td>\n",
       "      <td>...</td>\n",
       "      <td>6.590</td>\n",
       "      <td>6.42</td>\n",
       "      <td>5.56</td>\n",
       "      <td>4.800</td>\n",
       "      <td>5.910000</td>\n",
       "      <td>4.27</td>\n",
       "      <td>7.146667</td>\n",
       "      <td>9.355</td>\n",
       "      <td>6.203333</td>\n",
       "      <td>7.450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293603</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.15</td>\n",
       "      <td>4.76</td>\n",
       "      <td>3.880</td>\n",
       "      <td>4.025</td>\n",
       "      <td>5.11</td>\n",
       "      <td>3.460</td>\n",
       "      <td>3.97</td>\n",
       "      <td>5.90</td>\n",
       "      <td>4.14</td>\n",
       "      <td>...</td>\n",
       "      <td>6.305</td>\n",
       "      <td>6.58</td>\n",
       "      <td>5.51</td>\n",
       "      <td>4.700</td>\n",
       "      <td>5.971667</td>\n",
       "      <td>4.12</td>\n",
       "      <td>7.253333</td>\n",
       "      <td>9.470</td>\n",
       "      <td>6.290000</td>\n",
       "      <td>7.170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293604</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.56</td>\n",
       "      <td>4.51</td>\n",
       "      <td>4.210</td>\n",
       "      <td>4.080</td>\n",
       "      <td>4.97</td>\n",
       "      <td>3.565</td>\n",
       "      <td>4.12</td>\n",
       "      <td>5.59</td>\n",
       "      <td>4.14</td>\n",
       "      <td>...</td>\n",
       "      <td>6.465</td>\n",
       "      <td>6.29</td>\n",
       "      <td>5.34</td>\n",
       "      <td>4.695</td>\n",
       "      <td>6.031667</td>\n",
       "      <td>4.21</td>\n",
       "      <td>7.246667</td>\n",
       "      <td>9.725</td>\n",
       "      <td>6.233333</td>\n",
       "      <td>7.205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293605</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.78</td>\n",
       "      <td>4.58</td>\n",
       "      <td>4.070</td>\n",
       "      <td>3.995</td>\n",
       "      <td>4.88</td>\n",
       "      <td>3.520</td>\n",
       "      <td>3.97</td>\n",
       "      <td>5.39</td>\n",
       "      <td>4.04</td>\n",
       "      <td>...</td>\n",
       "      <td>6.950</td>\n",
       "      <td>7.29</td>\n",
       "      <td>5.61</td>\n",
       "      <td>4.795</td>\n",
       "      <td>5.920000</td>\n",
       "      <td>4.24</td>\n",
       "      <td>7.283333</td>\n",
       "      <td>9.585</td>\n",
       "      <td>6.236667</td>\n",
       "      <td>7.370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293606</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.34</td>\n",
       "      <td>4.39</td>\n",
       "      <td>4.035</td>\n",
       "      <td>4.105</td>\n",
       "      <td>5.16</td>\n",
       "      <td>3.470</td>\n",
       "      <td>4.11</td>\n",
       "      <td>5.74</td>\n",
       "      <td>3.95</td>\n",
       "      <td>...</td>\n",
       "      <td>6.070</td>\n",
       "      <td>5.94</td>\n",
       "      <td>5.40</td>\n",
       "      <td>4.840</td>\n",
       "      <td>5.998333</td>\n",
       "      <td>4.36</td>\n",
       "      <td>7.076667</td>\n",
       "      <td>9.730</td>\n",
       "      <td>6.296667</td>\n",
       "      <td>7.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293607</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.46</td>\n",
       "      <td>4.67</td>\n",
       "      <td>4.005</td>\n",
       "      <td>3.850</td>\n",
       "      <td>4.98</td>\n",
       "      <td>3.535</td>\n",
       "      <td>3.97</td>\n",
       "      <td>5.88</td>\n",
       "      <td>4.02</td>\n",
       "      <td>...</td>\n",
       "      <td>6.330</td>\n",
       "      <td>6.18</td>\n",
       "      <td>5.85</td>\n",
       "      <td>4.980</td>\n",
       "      <td>5.943333</td>\n",
       "      <td>4.11</td>\n",
       "      <td>7.190000</td>\n",
       "      <td>9.610</td>\n",
       "      <td>6.300000</td>\n",
       "      <td>7.515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293608</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.31</td>\n",
       "      <td>4.41</td>\n",
       "      <td>3.965</td>\n",
       "      <td>4.010</td>\n",
       "      <td>5.32</td>\n",
       "      <td>3.435</td>\n",
       "      <td>4.20</td>\n",
       "      <td>5.66</td>\n",
       "      <td>3.87</td>\n",
       "      <td>...</td>\n",
       "      <td>6.300</td>\n",
       "      <td>7.14</td>\n",
       "      <td>5.63</td>\n",
       "      <td>4.950</td>\n",
       "      <td>5.901667</td>\n",
       "      <td>4.32</td>\n",
       "      <td>7.076667</td>\n",
       "      <td>8.925</td>\n",
       "      <td>6.276667</td>\n",
       "      <td>7.430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293609</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.39</td>\n",
       "      <td>4.72</td>\n",
       "      <td>4.195</td>\n",
       "      <td>4.190</td>\n",
       "      <td>6.27</td>\n",
       "      <td>3.495</td>\n",
       "      <td>4.28</td>\n",
       "      <td>6.14</td>\n",
       "      <td>4.43</td>\n",
       "      <td>...</td>\n",
       "      <td>5.805</td>\n",
       "      <td>5.51</td>\n",
       "      <td>5.24</td>\n",
       "      <td>4.645</td>\n",
       "      <td>5.896667</td>\n",
       "      <td>4.63</td>\n",
       "      <td>6.603333</td>\n",
       "      <td>9.180</td>\n",
       "      <td>6.323333</td>\n",
       "      <td>7.090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293610</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.68</td>\n",
       "      <td>4.74</td>\n",
       "      <td>4.130</td>\n",
       "      <td>3.950</td>\n",
       "      <td>4.37</td>\n",
       "      <td>3.595</td>\n",
       "      <td>3.90</td>\n",
       "      <td>5.61</td>\n",
       "      <td>4.15</td>\n",
       "      <td>...</td>\n",
       "      <td>6.455</td>\n",
       "      <td>6.61</td>\n",
       "      <td>5.97</td>\n",
       "      <td>4.895</td>\n",
       "      <td>5.910000</td>\n",
       "      <td>4.23</td>\n",
       "      <td>7.386667</td>\n",
       "      <td>9.795</td>\n",
       "      <td>6.266667</td>\n",
       "      <td>7.085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293611</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.34</td>\n",
       "      <td>4.39</td>\n",
       "      <td>3.925</td>\n",
       "      <td>3.935</td>\n",
       "      <td>4.79</td>\n",
       "      <td>3.460</td>\n",
       "      <td>4.42</td>\n",
       "      <td>5.46</td>\n",
       "      <td>4.02</td>\n",
       "      <td>...</td>\n",
       "      <td>6.330</td>\n",
       "      <td>5.87</td>\n",
       "      <td>5.78</td>\n",
       "      <td>4.950</td>\n",
       "      <td>6.016667</td>\n",
       "      <td>3.95</td>\n",
       "      <td>7.276667</td>\n",
       "      <td>9.390</td>\n",
       "      <td>6.040000</td>\n",
       "      <td>7.370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293612</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.60</td>\n",
       "      <td>4.38</td>\n",
       "      <td>4.010</td>\n",
       "      <td>3.845</td>\n",
       "      <td>4.86</td>\n",
       "      <td>3.535</td>\n",
       "      <td>4.22</td>\n",
       "      <td>5.74</td>\n",
       "      <td>4.05</td>\n",
       "      <td>...</td>\n",
       "      <td>6.185</td>\n",
       "      <td>6.66</td>\n",
       "      <td>5.69</td>\n",
       "      <td>5.135</td>\n",
       "      <td>6.006667</td>\n",
       "      <td>4.02</td>\n",
       "      <td>7.106667</td>\n",
       "      <td>9.570</td>\n",
       "      <td>6.133333</td>\n",
       "      <td>7.460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293613</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.23</td>\n",
       "      <td>4.59</td>\n",
       "      <td>3.915</td>\n",
       "      <td>4.180</td>\n",
       "      <td>4.81</td>\n",
       "      <td>3.485</td>\n",
       "      <td>4.24</td>\n",
       "      <td>5.48</td>\n",
       "      <td>4.18</td>\n",
       "      <td>...</td>\n",
       "      <td>6.220</td>\n",
       "      <td>7.02</td>\n",
       "      <td>5.76</td>\n",
       "      <td>4.675</td>\n",
       "      <td>5.873333</td>\n",
       "      <td>3.94</td>\n",
       "      <td>7.110000</td>\n",
       "      <td>9.985</td>\n",
       "      <td>6.303333</td>\n",
       "      <td>7.270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293614</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.54</td>\n",
       "      <td>4.69</td>\n",
       "      <td>3.935</td>\n",
       "      <td>3.955</td>\n",
       "      <td>5.29</td>\n",
       "      <td>3.585</td>\n",
       "      <td>4.04</td>\n",
       "      <td>5.66</td>\n",
       "      <td>4.05</td>\n",
       "      <td>...</td>\n",
       "      <td>6.235</td>\n",
       "      <td>5.96</td>\n",
       "      <td>5.68</td>\n",
       "      <td>4.795</td>\n",
       "      <td>5.901667</td>\n",
       "      <td>4.41</td>\n",
       "      <td>7.113333</td>\n",
       "      <td>9.470</td>\n",
       "      <td>6.213333</td>\n",
       "      <td>7.260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293615</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.51</td>\n",
       "      <td>4.23</td>\n",
       "      <td>4.305</td>\n",
       "      <td>4.060</td>\n",
       "      <td>4.20</td>\n",
       "      <td>3.420</td>\n",
       "      <td>4.60</td>\n",
       "      <td>5.62</td>\n",
       "      <td>4.43</td>\n",
       "      <td>...</td>\n",
       "      <td>5.990</td>\n",
       "      <td>6.59</td>\n",
       "      <td>6.13</td>\n",
       "      <td>4.785</td>\n",
       "      <td>5.855000</td>\n",
       "      <td>4.52</td>\n",
       "      <td>6.450000</td>\n",
       "      <td>9.110</td>\n",
       "      <td>6.186667</td>\n",
       "      <td>7.310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293616</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.57</td>\n",
       "      <td>4.77</td>\n",
       "      <td>3.940</td>\n",
       "      <td>3.850</td>\n",
       "      <td>5.27</td>\n",
       "      <td>3.435</td>\n",
       "      <td>4.12</td>\n",
       "      <td>5.69</td>\n",
       "      <td>3.95</td>\n",
       "      <td>...</td>\n",
       "      <td>6.940</td>\n",
       "      <td>7.14</td>\n",
       "      <td>5.57</td>\n",
       "      <td>4.760</td>\n",
       "      <td>6.041667</td>\n",
       "      <td>4.04</td>\n",
       "      <td>7.260000</td>\n",
       "      <td>9.355</td>\n",
       "      <td>6.146667</td>\n",
       "      <td>7.315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293617</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.44</td>\n",
       "      <td>4.78</td>\n",
       "      <td>3.835</td>\n",
       "      <td>3.725</td>\n",
       "      <td>5.83</td>\n",
       "      <td>3.340</td>\n",
       "      <td>3.71</td>\n",
       "      <td>5.47</td>\n",
       "      <td>4.14</td>\n",
       "      <td>...</td>\n",
       "      <td>6.905</td>\n",
       "      <td>6.63</td>\n",
       "      <td>5.66</td>\n",
       "      <td>5.030</td>\n",
       "      <td>6.120000</td>\n",
       "      <td>3.99</td>\n",
       "      <td>7.416667</td>\n",
       "      <td>9.370</td>\n",
       "      <td>6.200000</td>\n",
       "      <td>7.410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293618</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.50</td>\n",
       "      <td>4.29</td>\n",
       "      <td>3.875</td>\n",
       "      <td>3.910</td>\n",
       "      <td>5.52</td>\n",
       "      <td>3.390</td>\n",
       "      <td>4.00</td>\n",
       "      <td>5.37</td>\n",
       "      <td>4.01</td>\n",
       "      <td>...</td>\n",
       "      <td>6.715</td>\n",
       "      <td>6.75</td>\n",
       "      <td>5.86</td>\n",
       "      <td>5.025</td>\n",
       "      <td>6.075000</td>\n",
       "      <td>4.16</td>\n",
       "      <td>7.513333</td>\n",
       "      <td>8.935</td>\n",
       "      <td>6.013333</td>\n",
       "      <td>7.640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293619</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.18</td>\n",
       "      <td>4.87</td>\n",
       "      <td>3.860</td>\n",
       "      <td>3.705</td>\n",
       "      <td>4.38</td>\n",
       "      <td>3.670</td>\n",
       "      <td>4.12</td>\n",
       "      <td>5.69</td>\n",
       "      <td>4.04</td>\n",
       "      <td>...</td>\n",
       "      <td>6.880</td>\n",
       "      <td>7.36</td>\n",
       "      <td>5.38</td>\n",
       "      <td>4.855</td>\n",
       "      <td>5.935000</td>\n",
       "      <td>4.41</td>\n",
       "      <td>7.443333</td>\n",
       "      <td>9.160</td>\n",
       "      <td>6.076667</td>\n",
       "      <td>7.595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293620</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.41</td>\n",
       "      <td>4.32</td>\n",
       "      <td>4.065</td>\n",
       "      <td>3.840</td>\n",
       "      <td>5.11</td>\n",
       "      <td>3.520</td>\n",
       "      <td>4.26</td>\n",
       "      <td>5.64</td>\n",
       "      <td>4.08</td>\n",
       "      <td>...</td>\n",
       "      <td>6.710</td>\n",
       "      <td>6.72</td>\n",
       "      <td>5.86</td>\n",
       "      <td>4.900</td>\n",
       "      <td>6.010000</td>\n",
       "      <td>3.99</td>\n",
       "      <td>7.456667</td>\n",
       "      <td>9.370</td>\n",
       "      <td>6.173333</td>\n",
       "      <td>7.515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293621</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.65</td>\n",
       "      <td>4.24</td>\n",
       "      <td>4.080</td>\n",
       "      <td>3.970</td>\n",
       "      <td>5.42</td>\n",
       "      <td>3.515</td>\n",
       "      <td>4.78</td>\n",
       "      <td>6.02</td>\n",
       "      <td>4.38</td>\n",
       "      <td>...</td>\n",
       "      <td>7.035</td>\n",
       "      <td>6.39</td>\n",
       "      <td>5.29</td>\n",
       "      <td>4.840</td>\n",
       "      <td>5.930000</td>\n",
       "      <td>4.18</td>\n",
       "      <td>6.993333</td>\n",
       "      <td>9.335</td>\n",
       "      <td>6.320000</td>\n",
       "      <td>7.225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293622</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.75</td>\n",
       "      <td>4.69</td>\n",
       "      <td>4.060</td>\n",
       "      <td>4.075</td>\n",
       "      <td>3.92</td>\n",
       "      <td>3.390</td>\n",
       "      <td>3.88</td>\n",
       "      <td>5.88</td>\n",
       "      <td>3.91</td>\n",
       "      <td>...</td>\n",
       "      <td>6.295</td>\n",
       "      <td>6.07</td>\n",
       "      <td>5.70</td>\n",
       "      <td>5.160</td>\n",
       "      <td>5.908333</td>\n",
       "      <td>4.15</td>\n",
       "      <td>7.333333</td>\n",
       "      <td>9.875</td>\n",
       "      <td>5.950000</td>\n",
       "      <td>7.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293623</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.43</td>\n",
       "      <td>4.63</td>\n",
       "      <td>4.045</td>\n",
       "      <td>3.790</td>\n",
       "      <td>6.42</td>\n",
       "      <td>3.595</td>\n",
       "      <td>4.04</td>\n",
       "      <td>5.68</td>\n",
       "      <td>3.79</td>\n",
       "      <td>...</td>\n",
       "      <td>6.340</td>\n",
       "      <td>6.54</td>\n",
       "      <td>5.34</td>\n",
       "      <td>4.895</td>\n",
       "      <td>6.123333</td>\n",
       "      <td>3.83</td>\n",
       "      <td>7.453333</td>\n",
       "      <td>9.465</td>\n",
       "      <td>6.276667</td>\n",
       "      <td>7.365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293624</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.68</td>\n",
       "      <td>4.52</td>\n",
       "      <td>4.120</td>\n",
       "      <td>4.000</td>\n",
       "      <td>4.39</td>\n",
       "      <td>3.520</td>\n",
       "      <td>4.34</td>\n",
       "      <td>5.78</td>\n",
       "      <td>4.00</td>\n",
       "      <td>...</td>\n",
       "      <td>6.320</td>\n",
       "      <td>6.26</td>\n",
       "      <td>5.49</td>\n",
       "      <td>4.920</td>\n",
       "      <td>5.921667</td>\n",
       "      <td>4.23</td>\n",
       "      <td>7.236667</td>\n",
       "      <td>9.825</td>\n",
       "      <td>6.166667</td>\n",
       "      <td>7.315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1293625</th>\n",
       "      <td>0.0</td>\n",
       "      <td>6.27</td>\n",
       "      <td>4.79</td>\n",
       "      <td>3.860</td>\n",
       "      <td>4.010</td>\n",
       "      <td>5.65</td>\n",
       "      <td>3.750</td>\n",
       "      <td>4.96</td>\n",
       "      <td>5.56</td>\n",
       "      <td>4.04</td>\n",
       "      <td>...</td>\n",
       "      <td>6.895</td>\n",
       "      <td>6.25</td>\n",
       "      <td>5.55</td>\n",
       "      <td>4.960</td>\n",
       "      <td>5.963333</td>\n",
       "      <td>4.31</td>\n",
       "      <td>6.880000</td>\n",
       "      <td>9.600</td>\n",
       "      <td>6.403333</td>\n",
       "      <td>7.230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34 rows × 17858 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Sickle Cell Anemia  A1BG  A1BG-AS1   A1CF    A2M  A2M-AS1  A2ML1  \\\n",
       "GSM1293592                 1.0  6.79      4.73  4.110  4.005     5.37  3.590   \n",
       "GSM1293593                 1.0  6.20      4.52  4.100  3.850     4.52  3.650   \n",
       "GSM1293594                 1.0  6.58      4.65  4.215  3.970     5.53  3.580   \n",
       "GSM1293595                 1.0  6.35      4.62  3.935  3.635     5.82  3.590   \n",
       "GSM1293596                 1.0  6.24      4.21  4.035  4.090     4.99  3.800   \n",
       "GSM1293597                 1.0  6.57      4.48  4.105  4.315     4.96  3.620   \n",
       "GSM1293598                 1.0  6.53      4.78  4.245  3.905     5.02  3.445   \n",
       "GSM1293599                 1.0  6.23      4.47  3.870  3.755     5.71  3.465   \n",
       "GSM1293600                 1.0  6.63      4.53  4.115  3.835     5.72  3.315   \n",
       "GSM1293601                 1.0  6.66      4.95  4.165  3.945     5.78  3.525   \n",
       "GSM1293602                 1.0  6.63      4.65  3.840  4.220     5.35  3.565   \n",
       "GSM1293603                 1.0  6.15      4.76  3.880  4.025     5.11  3.460   \n",
       "GSM1293604                 1.0  6.56      4.51  4.210  4.080     4.97  3.565   \n",
       "GSM1293605                 1.0  6.78      4.58  4.070  3.995     4.88  3.520   \n",
       "GSM1293606                 1.0  6.34      4.39  4.035  4.105     5.16  3.470   \n",
       "GSM1293607                 1.0  6.46      4.67  4.005  3.850     4.98  3.535   \n",
       "GSM1293608                 1.0  6.31      4.41  3.965  4.010     5.32  3.435   \n",
       "GSM1293609                 1.0  6.39      4.72  4.195  4.190     6.27  3.495   \n",
       "GSM1293610                 1.0  6.68      4.74  4.130  3.950     4.37  3.595   \n",
       "GSM1293611                 1.0  6.34      4.39  3.925  3.935     4.79  3.460   \n",
       "GSM1293612                 1.0  6.60      4.38  4.010  3.845     4.86  3.535   \n",
       "GSM1293613                 1.0  6.23      4.59  3.915  4.180     4.81  3.485   \n",
       "GSM1293614                 1.0  6.54      4.69  3.935  3.955     5.29  3.585   \n",
       "GSM1293615                 1.0  6.51      4.23  4.305  4.060     4.20  3.420   \n",
       "GSM1293616                 0.0  6.57      4.77  3.940  3.850     5.27  3.435   \n",
       "GSM1293617                 0.0  6.44      4.78  3.835  3.725     5.83  3.340   \n",
       "GSM1293618                 0.0  6.50      4.29  3.875  3.910     5.52  3.390   \n",
       "GSM1293619                 0.0  7.18      4.87  3.860  3.705     4.38  3.670   \n",
       "GSM1293620                 0.0  6.41      4.32  4.065  3.840     5.11  3.520   \n",
       "GSM1293621                 0.0  6.65      4.24  4.080  3.970     5.42  3.515   \n",
       "GSM1293622                 0.0  6.75      4.69  4.060  4.075     3.92  3.390   \n",
       "GSM1293623                 0.0  6.43      4.63  4.045  3.790     6.42  3.595   \n",
       "GSM1293624                 0.0  6.68      4.52  4.120  4.000     4.39  3.520   \n",
       "GSM1293625                 0.0  6.27      4.79  3.860  4.010     5.65  3.750   \n",
       "\n",
       "            A2MP1  A4GALT  A4GNT  ...  ZWILCH  ZWINT  ZXDA   ZXDB      ZXDC  \\\n",
       "GSM1293592   3.83    5.77   4.19  ...   6.510   6.63  5.66  4.870  5.723333   \n",
       "GSM1293593   4.23    5.86   4.15  ...   6.025   5.68  5.61  4.940  5.838333   \n",
       "GSM1293594   3.96    5.74   4.13  ...   6.210   6.39  5.63  4.865  5.936667   \n",
       "GSM1293595   3.94    5.48   4.00  ...   6.370   6.41  6.10  4.900  6.178333   \n",
       "GSM1293596   4.11    5.92   3.90  ...   6.135   6.52  5.60  4.790  6.088333   \n",
       "GSM1293597   3.77    5.67   4.18  ...   6.460   6.50  5.45  4.885  5.725000   \n",
       "GSM1293598   4.38    5.72   3.91  ...   5.245   5.98  5.40  4.785  6.140000   \n",
       "GSM1293599   4.29    5.31   3.71  ...   6.180   6.11  5.87  4.900  6.186667   \n",
       "GSM1293600   3.89    5.68   4.39  ...   6.640   6.78  6.04  5.000  5.970000   \n",
       "GSM1293601   3.99    5.92   4.01  ...   6.455   6.15  5.49  4.860  5.916667   \n",
       "GSM1293602   4.75    5.68   3.95  ...   6.590   6.42  5.56  4.800  5.910000   \n",
       "GSM1293603   3.97    5.90   4.14  ...   6.305   6.58  5.51  4.700  5.971667   \n",
       "GSM1293604   4.12    5.59   4.14  ...   6.465   6.29  5.34  4.695  6.031667   \n",
       "GSM1293605   3.97    5.39   4.04  ...   6.950   7.29  5.61  4.795  5.920000   \n",
       "GSM1293606   4.11    5.74   3.95  ...   6.070   5.94  5.40  4.840  5.998333   \n",
       "GSM1293607   3.97    5.88   4.02  ...   6.330   6.18  5.85  4.980  5.943333   \n",
       "GSM1293608   4.20    5.66   3.87  ...   6.300   7.14  5.63  4.950  5.901667   \n",
       "GSM1293609   4.28    6.14   4.43  ...   5.805   5.51  5.24  4.645  5.896667   \n",
       "GSM1293610   3.90    5.61   4.15  ...   6.455   6.61  5.97  4.895  5.910000   \n",
       "GSM1293611   4.42    5.46   4.02  ...   6.330   5.87  5.78  4.950  6.016667   \n",
       "GSM1293612   4.22    5.74   4.05  ...   6.185   6.66  5.69  5.135  6.006667   \n",
       "GSM1293613   4.24    5.48   4.18  ...   6.220   7.02  5.76  4.675  5.873333   \n",
       "GSM1293614   4.04    5.66   4.05  ...   6.235   5.96  5.68  4.795  5.901667   \n",
       "GSM1293615   4.60    5.62   4.43  ...   5.990   6.59  6.13  4.785  5.855000   \n",
       "GSM1293616   4.12    5.69   3.95  ...   6.940   7.14  5.57  4.760  6.041667   \n",
       "GSM1293617   3.71    5.47   4.14  ...   6.905   6.63  5.66  5.030  6.120000   \n",
       "GSM1293618   4.00    5.37   4.01  ...   6.715   6.75  5.86  5.025  6.075000   \n",
       "GSM1293619   4.12    5.69   4.04  ...   6.880   7.36  5.38  4.855  5.935000   \n",
       "GSM1293620   4.26    5.64   4.08  ...   6.710   6.72  5.86  4.900  6.010000   \n",
       "GSM1293621   4.78    6.02   4.38  ...   7.035   6.39  5.29  4.840  5.930000   \n",
       "GSM1293622   3.88    5.88   3.91  ...   6.295   6.07  5.70  5.160  5.908333   \n",
       "GSM1293623   4.04    5.68   3.79  ...   6.340   6.54  5.34  4.895  6.123333   \n",
       "GSM1293624   4.34    5.78   4.00  ...   6.320   6.26  5.49  4.920  5.921667   \n",
       "GSM1293625   4.96    5.56   4.04  ...   6.895   6.25  5.55  4.960  5.963333   \n",
       "\n",
       "            ZYG11A    ZYG11B    ZYX     ZZEF1   ZZZ3  \n",
       "GSM1293592    4.32  7.133333  8.890  5.856667  7.495  \n",
       "GSM1293593    4.37  7.016667  9.260  6.500000  7.055  \n",
       "GSM1293594    4.23  6.723333  9.435  6.313333  7.050  \n",
       "GSM1293595    4.19  7.323333  9.610  6.313333  7.485  \n",
       "GSM1293596    4.28  6.870000  9.210  6.313333  6.985  \n",
       "GSM1293597    4.27  7.216667  9.730  6.206667  7.305  \n",
       "GSM1293598    4.31  7.013333  9.175  6.386667  7.110  \n",
       "GSM1293599    4.23  6.983333  9.150  6.343333  7.205  \n",
       "GSM1293600    4.27  7.260000  8.775  5.833333  7.585  \n",
       "GSM1293601    4.17  7.166667  9.240  6.220000  7.320  \n",
       "GSM1293602    4.27  7.146667  9.355  6.203333  7.450  \n",
       "GSM1293603    4.12  7.253333  9.470  6.290000  7.170  \n",
       "GSM1293604    4.21  7.246667  9.725  6.233333  7.205  \n",
       "GSM1293605    4.24  7.283333  9.585  6.236667  7.370  \n",
       "GSM1293606    4.36  7.076667  9.730  6.296667  7.160  \n",
       "GSM1293607    4.11  7.190000  9.610  6.300000  7.515  \n",
       "GSM1293608    4.32  7.076667  8.925  6.276667  7.430  \n",
       "GSM1293609    4.63  6.603333  9.180  6.323333  7.090  \n",
       "GSM1293610    4.23  7.386667  9.795  6.266667  7.085  \n",
       "GSM1293611    3.95  7.276667  9.390  6.040000  7.370  \n",
       "GSM1293612    4.02  7.106667  9.570  6.133333  7.460  \n",
       "GSM1293613    3.94  7.110000  9.985  6.303333  7.270  \n",
       "GSM1293614    4.41  7.113333  9.470  6.213333  7.260  \n",
       "GSM1293615    4.52  6.450000  9.110  6.186667  7.310  \n",
       "GSM1293616    4.04  7.260000  9.355  6.146667  7.315  \n",
       "GSM1293617    3.99  7.416667  9.370  6.200000  7.410  \n",
       "GSM1293618    4.16  7.513333  8.935  6.013333  7.640  \n",
       "GSM1293619    4.41  7.443333  9.160  6.076667  7.595  \n",
       "GSM1293620    3.99  7.456667  9.370  6.173333  7.515  \n",
       "GSM1293621    4.18  6.993333  9.335  6.320000  7.225  \n",
       "GSM1293622    4.15  7.333333  9.875  5.950000  7.160  \n",
       "GSM1293623    3.83  7.453333  9.465  6.276667  7.365  \n",
       "GSM1293624    4.23  7.236667  9.825  6.166667  7.315  \n",
       "GSM1293625    4.31  6.880000  9.600  6.403333  7.230  \n",
       "\n",
       "[34 rows x 17858 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 34 samples.\n",
      "For the feature 'Sickle Cell Anemia', the least common label is '0.0' with 10 occurrences. This represents 29.41% of the dataset.\n",
      "The distribution of the feature 'Sickle Cell Anemia' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_available = False\n",
    "\n",
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
