{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = 'Testicular Cancer'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:21.140790869Z",
     "start_time": "2024-02-27T00:22:21.133208795Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Cervical_Cancer_(CESC)',\n",
       " 'TCGA_Kidney_Chromophobe_(KICH)',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)',\n",
       " 'TCGA_Lower_Grade_Glioma_(LGG)',\n",
       " 'TCGA_Melanoma_(SKCM)',\n",
       " 'TCGA_Mesothelioma_(MESO)',\n",
       " 'TCGA_Testicular_Cancer_(TGCT)',\n",
       " 'TCGA_Uterine_Carcinosarcoma_(UCS)']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:21.985144212Z",
     "start_time": "2024-02-27T00:22:21.981084968Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\TCGA\\\\TCGA_Testicular_Cancer_(TGCT)\\\\TCGA.TGCT.sampleMap_TGCT_clinicalMatrix',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\TCGA\\\\TCGA_Testicular_Cancer_(TGCT)\\\\TCGA.TGCT.sampleMap_HiSeqV2_PANCAN.gz')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_subdir = \"TCGA_Testicular_Cancer_(TGCT)\"\n",
    "cohort = 'Xena'\n",
    "trait_type = 'binary'\n",
    "is_available = True\n",
    "\n",
    "cohort_dir = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "clinical_data_file, genetic_data_file = xena_get_relevant_filepaths(cohort_dir)\n",
    "clinical_data_file, genetic_data_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.794819905Z",
     "start_time": "2024-02-27T00:22:22.357077391Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.798785383Z",
     "start_time": "2024-02-27T00:22:25.796474039Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'age_at_initial_pathologic_diagnosis',\n",
       " 'bcr_followup_barcode',\n",
       " 'bcr_patient_barcode',\n",
       " 'bcr_sample_barcode',\n",
       " 'bilateral_diagnosis_timing_type']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.802838875Z",
     "start_time": "2024-02-27T00:22:25.798659327Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'age_at_initial_pathologic_diagnosis', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'bilateral_diagnosis_timing_type', 'clinical_M', 'clinical_N', 'clinical_T', 'clinical_stage', 'days_to_bilateral_tumor_dx', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_after_initial_treatment', 'days_to_post_orchi_serum_test', 'days_to_pre_orchi_serum_test', 'eastern_cancer_oncology_group', 'family_history_other_cancer', 'family_history_testicular_cancer', 'family_member_relationship_type', 'first_treatment_success', 'form_completion_date', 'gender', 'histological_percentage', 'histological_type', 'history_fertility', 'history_hypospadias', 'history_of_neoadjuvant_treatment', 'history_of_undescended_testis', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'igcccg_stage', 'informed_consent_verified', 'init_pathology_dx_method_other', 'initial_pathologic_diagnosis_method', 'initial_weight', 'intratubular_germ_cell_neoplasm', 'is_ffpe', 'karnofsky_performance_score', 'laterality', 'level_of_non_descent', 'lost_follow_up', 'lymphovascular_invasion_present', 'molecular_test_result', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_after_initial_treatment', 'oct_embedded', 'other_dx', 'pathologic_M', 'pathologic_N', 'pathologic_T', 'pathologic_stage', 'pathology_report_file_name', 'patient_death_reason', 'patient_id', 'person_neoplasm_cancer_status', 'post_orchi_afp', 'post_orchi_hcg', 'post_orchi_ldh', 'post_orchi_lh', 'post_orchi_lymph_node_dissection', 'post_orchi_testosterone', 'postoperative_rx_tx', 'postoperative_tx', 'pre_orchi_afp', 'pre_orchi_hcg', 'pre_orchi_ldh', 'pre_orchi_lh', 'pre_orchi_testosterone', 'primary_therapy_outcome_success', 'radiation_therapy', 'relation_testicular_cancer', 'relative_family_cancer_hx_text', 'sample_type', 'sample_type_id', 'serum_markers', 'source_of_patient_death_reason', 'synchronous_tumor_histology_pct', 'synchronous_tumor_histology_type', 'system_version', 'testis_tumor_macroextent', 'testis_tumor_macroextent_other', 'testis_tumor_microextent', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'tumor_tissue_site', 'undescended_testis_corrected', 'undescended_testis_corrected_age', 'undescended_testis_method_left', 'undescended_testis_method_right', 'vial_number', 'vital_status', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_TGCT_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_TGCT_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_TGCT_mutation_ucsc_maf_gene', '_GENOMIC_ID_TCGA_TGCT_exp_HiSeqV2', '_GENOMIC_ID_TCGA_TGCT_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_TGCT_hMethyl450', '_GENOMIC_ID_TCGA_TGCT_gistic2', '_GENOMIC_ID_data/public/TCGA/TGCT/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_TGCT_gistic2thd', '_GENOMIC_ID_TCGA_TGCT_mutation_bcm_gene', '_GENOMIC_ID_TCGA_TGCT_miRNA_HiSeq', '_GENOMIC_ID_TCGA_TGCT_mutation_broad_gene', '_GENOMIC_ID_TCGA_TGCT_PDMRNAseq', '_GENOMIC_ID_TCGA_TGCT_RPPA', '_GENOMIC_ID_TCGA_TGCT_exp_HiSeqV2_percentile', '_GENOMIC_ID_TCGA_TGCT_mutation_bcgsc_gene']\\n\""
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.864092575Z",
     "start_time": "2024-02-27T00:22:25.802001989Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = ['age_at_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = ['gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "\n",
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:26.108535524Z",
     "start_time": "2024-02-27T00:22:26.101435803Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'\n",
    "gender_col = 'gender'\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.602458026Z",
     "start_time": "2024-02-27T00:22:27.839636769Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-2G-AAHP-01</th>\n",
       "      <th>TCGA-4K-AAAL-01</th>\n",
       "      <th>TCGA-2G-AALP-01</th>\n",
       "      <th>TCGA-2G-AAGI-05</th>\n",
       "      <th>TCGA-ZM-AA0N-01</th>\n",
       "      <th>TCGA-2G-AAGI-01</th>\n",
       "      <th>TCGA-2G-AALT-01</th>\n",
       "      <th>TCGA-2G-AAKO-01</th>\n",
       "      <th>TCGA-2G-AAFG-01</th>\n",
       "      <th>TCGA-2G-AAG6-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-2G-AAKG-05</th>\n",
       "      <th>TCGA-2G-AAG9-01</th>\n",
       "      <th>TCGA-XE-AANR-01</th>\n",
       "      <th>TCGA-YU-A94I-01</th>\n",
       "      <th>TCGA-2G-AAFO-01</th>\n",
       "      <th>TCGA-2G-AALR-01</th>\n",
       "      <th>TCGA-SN-A84X-01</th>\n",
       "      <th>TCGA-XE-A8H5-01</th>\n",
       "      <th>TCGA-XE-AAOD-01</th>\n",
       "      <th>TCGA-2G-AAHT-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <td>-1.064292</td>\n",
       "      <td>-0.550492</td>\n",
       "      <td>-0.357592</td>\n",
       "      <td>-2.170992</td>\n",
       "      <td>-1.099292</td>\n",
       "      <td>0.287308</td>\n",
       "      <td>1.117908</td>\n",
       "      <td>-1.294892</td>\n",
       "      <td>-0.312792</td>\n",
       "      <td>0.377908</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.620992</td>\n",
       "      <td>-0.223592</td>\n",
       "      <td>-1.464992</td>\n",
       "      <td>-0.650692</td>\n",
       "      <td>-0.680692</td>\n",
       "      <td>0.141308</td>\n",
       "      <td>0.086308</td>\n",
       "      <td>-1.450792</td>\n",
       "      <td>-1.183792</td>\n",
       "      <td>-0.825592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIF3A</th>\n",
       "      <td>2.004674</td>\n",
       "      <td>2.885874</td>\n",
       "      <td>3.188374</td>\n",
       "      <td>1.747974</td>\n",
       "      <td>1.870874</td>\n",
       "      <td>5.475574</td>\n",
       "      <td>4.570574</td>\n",
       "      <td>0.672174</td>\n",
       "      <td>0.712774</td>\n",
       "      <td>4.982474</td>\n",
       "      <td>...</td>\n",
       "      <td>1.884474</td>\n",
       "      <td>2.449574</td>\n",
       "      <td>1.259574</td>\n",
       "      <td>2.959974</td>\n",
       "      <td>1.472474</td>\n",
       "      <td>4.936574</td>\n",
       "      <td>4.065074</td>\n",
       "      <td>1.933374</td>\n",
       "      <td>1.659474</td>\n",
       "      <td>0.928274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>1.875265</td>\n",
       "      <td>9.089365</td>\n",
       "      <td>0.116965</td>\n",
       "      <td>9.729765</td>\n",
       "      <td>9.669065</td>\n",
       "      <td>0.576865</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>9.999865</td>\n",
       "      <td>9.417865</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>...</td>\n",
       "      <td>2.382765</td>\n",
       "      <td>9.357565</td>\n",
       "      <td>9.222665</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>7.320565</td>\n",
       "      <td>0.316965</td>\n",
       "      <td>3.934365</td>\n",
       "      <td>9.994165</td>\n",
       "      <td>8.364465</td>\n",
       "      <td>9.605365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF10</th>\n",
       "      <td>-0.002472</td>\n",
       "      <td>0.790228</td>\n",
       "      <td>0.375328</td>\n",
       "      <td>0.640328</td>\n",
       "      <td>0.777128</td>\n",
       "      <td>-0.097272</td>\n",
       "      <td>-0.082872</td>\n",
       "      <td>0.591028</td>\n",
       "      <td>-0.105672</td>\n",
       "      <td>0.197328</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.256172</td>\n",
       "      <td>0.515428</td>\n",
       "      <td>0.598928</td>\n",
       "      <td>-0.035872</td>\n",
       "      <td>0.400628</td>\n",
       "      <td>0.221928</td>\n",
       "      <td>0.298228</td>\n",
       "      <td>0.427928</td>\n",
       "      <td>0.438728</td>\n",
       "      <td>0.413928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF11</th>\n",
       "      <td>-0.615978</td>\n",
       "      <td>0.056122</td>\n",
       "      <td>-0.587578</td>\n",
       "      <td>0.514822</td>\n",
       "      <td>-0.449078</td>\n",
       "      <td>-0.798378</td>\n",
       "      <td>-0.413378</td>\n",
       "      <td>-0.223078</td>\n",
       "      <td>0.662422</td>\n",
       "      <td>-0.389178</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.825278</td>\n",
       "      <td>0.179822</td>\n",
       "      <td>-0.506078</td>\n",
       "      <td>-0.504578</td>\n",
       "      <td>-0.595578</td>\n",
       "      <td>0.047122</td>\n",
       "      <td>-0.107178</td>\n",
       "      <td>-0.019278</td>\n",
       "      <td>-0.180378</td>\n",
       "      <td>-0.427978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNGT1</th>\n",
       "      <td>0.104310</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>0.152010</td>\n",
       "      <td>-0.837090</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>0.877510</td>\n",
       "      <td>0.715610</td>\n",
       "      <td>0.145810</td>\n",
       "      <td>-0.480790</td>\n",
       "      <td>0.644310</td>\n",
       "      <td>...</td>\n",
       "      <td>0.572410</td>\n",
       "      <td>-0.707890</td>\n",
       "      <td>-0.303290</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>0.207210</td>\n",
       "      <td>-0.433390</td>\n",
       "      <td>0.024010</td>\n",
       "      <td>0.781010</td>\n",
       "      <td>0.009010</td>\n",
       "      <td>-0.744090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TULP3</th>\n",
       "      <td>2.237623</td>\n",
       "      <td>1.823623</td>\n",
       "      <td>2.064423</td>\n",
       "      <td>2.454823</td>\n",
       "      <td>2.661123</td>\n",
       "      <td>1.331623</td>\n",
       "      <td>1.351723</td>\n",
       "      <td>2.375923</td>\n",
       "      <td>1.729723</td>\n",
       "      <td>2.133523</td>\n",
       "      <td>...</td>\n",
       "      <td>0.697823</td>\n",
       "      <td>3.338823</td>\n",
       "      <td>2.536723</td>\n",
       "      <td>1.265223</td>\n",
       "      <td>1.816023</td>\n",
       "      <td>1.164723</td>\n",
       "      <td>1.600223</td>\n",
       "      <td>3.572423</td>\n",
       "      <td>1.750023</td>\n",
       "      <td>3.034123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL6B</th>\n",
       "      <td>-0.107927</td>\n",
       "      <td>0.450673</td>\n",
       "      <td>-1.355327</td>\n",
       "      <td>0.676473</td>\n",
       "      <td>-0.923627</td>\n",
       "      <td>1.177673</td>\n",
       "      <td>2.341973</td>\n",
       "      <td>0.678373</td>\n",
       "      <td>-0.196027</td>\n",
       "      <td>0.711873</td>\n",
       "      <td>...</td>\n",
       "      <td>0.482273</td>\n",
       "      <td>0.308173</td>\n",
       "      <td>0.226273</td>\n",
       "      <td>0.562773</td>\n",
       "      <td>0.612773</td>\n",
       "      <td>0.253773</td>\n",
       "      <td>0.755673</td>\n",
       "      <td>-0.254427</td>\n",
       "      <td>0.718673</td>\n",
       "      <td>2.040473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSTK1</th>\n",
       "      <td>-1.888995</td>\n",
       "      <td>-0.786595</td>\n",
       "      <td>-0.779395</td>\n",
       "      <td>-1.329995</td>\n",
       "      <td>-0.866495</td>\n",
       "      <td>-0.688895</td>\n",
       "      <td>0.415005</td>\n",
       "      <td>-0.762395</td>\n",
       "      <td>-1.205195</td>\n",
       "      <td>-0.357695</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.163395</td>\n",
       "      <td>-1.039795</td>\n",
       "      <td>-1.417595</td>\n",
       "      <td>-1.395195</td>\n",
       "      <td>-0.661895</td>\n",
       "      <td>-1.282895</td>\n",
       "      <td>-0.642795</td>\n",
       "      <td>-2.410995</td>\n",
       "      <td>-0.671495</td>\n",
       "      <td>-1.830395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELP</th>\n",
       "      <td>0.116167</td>\n",
       "      <td>-0.403633</td>\n",
       "      <td>-3.387433</td>\n",
       "      <td>1.166867</td>\n",
       "      <td>-1.060333</td>\n",
       "      <td>0.702067</td>\n",
       "      <td>0.277967</td>\n",
       "      <td>0.600367</td>\n",
       "      <td>-0.053133</td>\n",
       "      <td>2.017067</td>\n",
       "      <td>...</td>\n",
       "      <td>0.438867</td>\n",
       "      <td>-0.330533</td>\n",
       "      <td>-0.358933</td>\n",
       "      <td>-3.452133</td>\n",
       "      <td>1.554067</td>\n",
       "      <td>-5.040133</td>\n",
       "      <td>1.267167</td>\n",
       "      <td>-0.491333</td>\n",
       "      <td>1.416267</td>\n",
       "      <td>0.174667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17339 rows × 156 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TCGA-2G-AAHP-01  TCGA-4K-AAAL-01  TCGA-2G-AALP-01  TCGA-2G-AAGI-05  \\\n",
       "ARHGEF10L        -1.064292        -0.550492        -0.357592        -2.170992   \n",
       "HIF3A             2.004674         2.885874         3.188374         1.747974   \n",
       "RNF17             1.875265         9.089365         0.116965         9.729765   \n",
       "RNF10            -0.002472         0.790228         0.375328         0.640328   \n",
       "RNF11            -0.615978         0.056122        -0.587578         0.514822   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             0.104310        -1.281390         0.152010        -0.837090   \n",
       "TULP3             2.237623         1.823623         2.064423         2.454823   \n",
       "BCL6B            -0.107927         0.450673        -1.355327         0.676473   \n",
       "GSTK1            -1.888995        -0.786595        -0.779395        -1.329995   \n",
       "SELP              0.116167        -0.403633        -3.387433         1.166867   \n",
       "\n",
       "           TCGA-ZM-AA0N-01  TCGA-2G-AAGI-01  TCGA-2G-AALT-01  TCGA-2G-AAKO-01  \\\n",
       "ARHGEF10L        -1.099292         0.287308         1.117908        -1.294892   \n",
       "HIF3A             1.870874         5.475574         4.570574         0.672174   \n",
       "RNF17             9.669065         0.576865        -0.531035         9.999865   \n",
       "RNF10             0.777128        -0.097272        -0.082872         0.591028   \n",
       "RNF11            -0.449078        -0.798378        -0.413378        -0.223078   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -1.281390         0.877510         0.715610         0.145810   \n",
       "TULP3             2.661123         1.331623         1.351723         2.375923   \n",
       "BCL6B            -0.923627         1.177673         2.341973         0.678373   \n",
       "GSTK1            -0.866495        -0.688895         0.415005        -0.762395   \n",
       "SELP             -1.060333         0.702067         0.277967         0.600367   \n",
       "\n",
       "           TCGA-2G-AAFG-01  TCGA-2G-AAG6-01  ...  TCGA-2G-AAKG-05  \\\n",
       "ARHGEF10L        -0.312792         0.377908  ...        -1.620992   \n",
       "HIF3A             0.712774         4.982474  ...         1.884474   \n",
       "RNF17             9.417865        -0.531035  ...         2.382765   \n",
       "RNF10            -0.105672         0.197328  ...        -0.256172   \n",
       "RNF11             0.662422        -0.389178  ...        -0.825278   \n",
       "...                    ...              ...  ...              ...   \n",
       "GNGT1            -0.480790         0.644310  ...         0.572410   \n",
       "TULP3             1.729723         2.133523  ...         0.697823   \n",
       "BCL6B            -0.196027         0.711873  ...         0.482273   \n",
       "GSTK1            -1.205195        -0.357695  ...        -0.163395   \n",
       "SELP             -0.053133         2.017067  ...         0.438867   \n",
       "\n",
       "           TCGA-2G-AAG9-01  TCGA-XE-AANR-01  TCGA-YU-A94I-01  TCGA-2G-AAFO-01  \\\n",
       "ARHGEF10L        -0.223592        -1.464992        -0.650692        -0.680692   \n",
       "HIF3A             2.449574         1.259574         2.959974         1.472474   \n",
       "RNF17             9.357565         9.222665        -0.531035         7.320565   \n",
       "RNF10             0.515428         0.598928        -0.035872         0.400628   \n",
       "RNF11             0.179822        -0.506078        -0.504578        -0.595578   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -0.707890        -0.303290        -1.281390         0.207210   \n",
       "TULP3             3.338823         2.536723         1.265223         1.816023   \n",
       "BCL6B             0.308173         0.226273         0.562773         0.612773   \n",
       "GSTK1            -1.039795        -1.417595        -1.395195        -0.661895   \n",
       "SELP             -0.330533        -0.358933        -3.452133         1.554067   \n",
       "\n",
       "           TCGA-2G-AALR-01  TCGA-SN-A84X-01  TCGA-XE-A8H5-01  TCGA-XE-AAOD-01  \\\n",
       "ARHGEF10L         0.141308         0.086308        -1.450792        -1.183792   \n",
       "HIF3A             4.936574         4.065074         1.933374         1.659474   \n",
       "RNF17             0.316965         3.934365         9.994165         8.364465   \n",
       "RNF10             0.221928         0.298228         0.427928         0.438728   \n",
       "RNF11             0.047122        -0.107178        -0.019278        -0.180378   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -0.433390         0.024010         0.781010         0.009010   \n",
       "TULP3             1.164723         1.600223         3.572423         1.750023   \n",
       "BCL6B             0.253773         0.755673        -0.254427         0.718673   \n",
       "GSTK1            -1.282895        -0.642795        -2.410995        -0.671495   \n",
       "SELP             -5.040133         1.267167        -0.491333         1.416267   \n",
       "\n",
       "           TCGA-2G-AAHT-01  \n",
       "ARHGEF10L        -0.825592  \n",
       "HIF3A             0.928274  \n",
       "RNF17             9.605365  \n",
       "RNF10             0.413928  \n",
       "RNF11            -0.427978  \n",
       "...                    ...  \n",
       "GNGT1            -0.744090  \n",
       "TULP3             3.034123  \n",
       "BCL6B             2.040473  \n",
       "GSTK1            -1.830395  \n",
       "SELP              0.174667  \n",
       "\n",
       "[17339 rows x 156 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.720144304Z",
     "start_time": "2024-02-27T00:23:16.643071320Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Testicular Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAEW-01</th>\n",
       "      <td>1</td>\n",
       "      <td>31.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.247192</td>\n",
       "      <td>1.370174</td>\n",
       "      <td>9.255665</td>\n",
       "      <td>0.322728</td>\n",
       "      <td>-0.485378</td>\n",
       "      <td>-0.61291</td>\n",
       "      <td>1.315306</td>\n",
       "      <td>...</td>\n",
       "      <td>3.498714</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>4.207622</td>\n",
       "      <td>-0.237217</td>\n",
       "      <td>1.597767</td>\n",
       "      <td>0.23401</td>\n",
       "      <td>2.696223</td>\n",
       "      <td>3.097773</td>\n",
       "      <td>-1.468595</td>\n",
       "      <td>-0.043933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAEX-01</th>\n",
       "      <td>1</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.743692</td>\n",
       "      <td>1.814274</td>\n",
       "      <td>9.680665</td>\n",
       "      <td>0.636028</td>\n",
       "      <td>0.108222</td>\n",
       "      <td>-1.26731</td>\n",
       "      <td>1.206606</td>\n",
       "      <td>...</td>\n",
       "      <td>3.790214</td>\n",
       "      <td>0.487618</td>\n",
       "      <td>5.504822</td>\n",
       "      <td>0.846083</td>\n",
       "      <td>0.797867</td>\n",
       "      <td>0.02141</td>\n",
       "      <td>2.778623</td>\n",
       "      <td>0.329973</td>\n",
       "      <td>-1.447895</td>\n",
       "      <td>1.108067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAF1-01</th>\n",
       "      <td>1</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.397192</td>\n",
       "      <td>1.187274</td>\n",
       "      <td>9.598765</td>\n",
       "      <td>0.838328</td>\n",
       "      <td>-0.096978</td>\n",
       "      <td>-1.04111</td>\n",
       "      <td>1.560006</td>\n",
       "      <td>...</td>\n",
       "      <td>3.421714</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>4.813122</td>\n",
       "      <td>-0.005417</td>\n",
       "      <td>1.629067</td>\n",
       "      <td>0.64101</td>\n",
       "      <td>2.337123</td>\n",
       "      <td>2.008273</td>\n",
       "      <td>-1.256195</td>\n",
       "      <td>0.153367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAF4-01</th>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.994992</td>\n",
       "      <td>1.597874</td>\n",
       "      <td>9.971765</td>\n",
       "      <td>0.339728</td>\n",
       "      <td>-0.098878</td>\n",
       "      <td>-1.05091</td>\n",
       "      <td>0.858706</td>\n",
       "      <td>...</td>\n",
       "      <td>4.097814</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>4.213022</td>\n",
       "      <td>2.064083</td>\n",
       "      <td>1.162567</td>\n",
       "      <td>0.53161</td>\n",
       "      <td>3.066823</td>\n",
       "      <td>-0.735627</td>\n",
       "      <td>-1.617995</td>\n",
       "      <td>-0.710533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2G-AAF6-01</th>\n",
       "      <td>1</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.498792</td>\n",
       "      <td>2.274374</td>\n",
       "      <td>8.821565</td>\n",
       "      <td>0.247628</td>\n",
       "      <td>0.408722</td>\n",
       "      <td>-0.65491</td>\n",
       "      <td>1.862706</td>\n",
       "      <td>...</td>\n",
       "      <td>4.072914</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>5.124822</td>\n",
       "      <td>1.812383</td>\n",
       "      <td>0.734067</td>\n",
       "      <td>-0.88259</td>\n",
       "      <td>2.702423</td>\n",
       "      <td>-1.043427</td>\n",
       "      <td>-1.584895</td>\n",
       "      <td>-0.367333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Testicular Cancer   Age  Gender  ARHGEF10L     HIF3A  \\\n",
       "sampleID                                                                \n",
       "TCGA-2G-AAEW-01                  1  31.0     1.0  -1.247192  1.370174   \n",
       "TCGA-2G-AAEX-01                  1  38.0     1.0  -1.743692  1.814274   \n",
       "TCGA-2G-AAF1-01                  1  28.0     1.0  -1.397192  1.187274   \n",
       "TCGA-2G-AAF4-01                  1  30.0     1.0  -0.994992  1.597874   \n",
       "TCGA-2G-AAF6-01                  1  28.0     1.0  -0.498792  2.274374   \n",
       "\n",
       "                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n",
       "sampleID                                                          ...   \n",
       "TCGA-2G-AAEW-01  9.255665  0.322728 -0.485378 -0.61291  1.315306  ...   \n",
       "TCGA-2G-AAEX-01  9.680665  0.636028  0.108222 -1.26731  1.206606  ...   \n",
       "TCGA-2G-AAF1-01  9.598765  0.838328 -0.096978 -1.04111  1.560006  ...   \n",
       "TCGA-2G-AAF4-01  9.971765  0.339728 -0.098878 -1.05091  0.858706  ...   \n",
       "TCGA-2G-AAF6-01  8.821565  0.247628  0.408722 -0.65491  1.862706  ...   \n",
       "\n",
       "                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n",
       "sampleID                                                                     \n",
       "TCGA-2G-AAEW-01  3.498714 -0.086682  4.207622 -0.237217  1.597767  0.23401   \n",
       "TCGA-2G-AAEX-01  3.790214  0.487618  5.504822  0.846083  0.797867  0.02141   \n",
       "TCGA-2G-AAF1-01  3.421714 -0.086682  4.813122 -0.005417  1.629067  0.64101   \n",
       "TCGA-2G-AAF4-01  4.097814 -0.086682  4.213022  2.064083  1.162567  0.53161   \n",
       "TCGA-2G-AAF6-01  4.072914 -0.086682  5.124822  1.812383  0.734067 -0.88259   \n",
       "\n",
       "                    TULP3     BCL6B     GSTK1      SELP  \n",
       "sampleID                                                 \n",
       "TCGA-2G-AAEW-01  2.696223  3.097773 -1.468595 -0.043933  \n",
       "TCGA-2G-AAEX-01  2.778623  0.329973 -1.447895  1.108067  \n",
       "TCGA-2G-AAF1-01  2.337123  2.008273 -1.256195  0.153367  \n",
       "TCGA-2G-AAF4-01  3.066823 -0.735627 -1.617995 -0.710533  \n",
       "TCGA-2G-AAF6-01  2.702423 -1.043427 -1.584895 -0.367333  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.731303752Z",
     "start_time": "2024-02-27T00:23:16.717316474Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 139 samples.\n",
      "For the feature 'Testicular Cancer', the least common label is '1' with 139 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Testicular Cancer' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 26.0\n",
      "  50% (Median): 31.0\n",
      "  75%: 37.0\n",
      "Min: 14.0\n",
      "Max: 67.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 139 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.252387093Z",
     "start_time": "2024-02-27T00:23:16.722533204Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302271525Z",
     "start_time": "2024-02-27T00:23:30.295066253Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output2\\Jiayi\\Testicular-Cancer\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE12630',\n",
       " 'GSE1818',\n",
       " 'GSE28094',\n",
       " 'GSE3921',\n",
       " 'GSE42647',\n",
       " 'GSE59520',\n",
       " 'GSE62523']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Testicular-Cancer\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE12630\\\\GSE12630_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE12630\\\\GSE12630-GPL7280_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE12630\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE1818\\\\GSE1818_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE1818\\\\GSE1818_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Empty\n",
    "cohort = accession_num = \"GSE1818\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE28094\\\\GSE28094_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE28094\\\\GSE28094_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Biased\n",
    "cohort = accession_num = \"GSE28094\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE3921\\\\GSE3921_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE3921\\\\GSE3921-GPL1283_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No trait information\n",
    "cohort = accession_num = \"GSE3921\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE42647\\\\GSE42647_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE42647\\\\GSE42647-GPL10558_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No trait information\n",
    "cohort = accession_num = \"GSE42647\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE59520\\\\GSE59520_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE59520\\\\GSE59520_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE59520\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE62523\\\\GSE62523_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Testicular-Cancer\\\\GSE62523\\\\GSE62523_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE62523\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"Gene expression profiles of HMEC-1 after exposure to the chemotherapeutic drugs bleomycin and cisplatin with untreated samples as control\"\n",
      "!Series_summary\t\"Chemotherapy-related endothelial damage contributes to the early development of cardiovascular morbidity in testicular cancer patients. We aimed to identify relevant mechanisms of and search for candidate biomarkers for this endothelial damage.\"\n",
      "!Series_summary\t\"Human micro-vascular endothelial cells (HMEC-1) were exposed to bleomycin or cisplatin with untreated samples as control. 18k cDNA microarrays were used. Gene expression differences were analysed at single gene level and in gene sets clustered in biological pathways and validated by qRT-PCR. Protein levels of a candidate biomarker were measured in testicular cancer patient plasma before, during and after bleomycin-etoposide-cisplatin chemotherapy, and related to endothelial damage biomarkers (von Willebrand Factor (vWF), high-sensitivity C-Reactive Protein (hsCRP)).\"\n",
      "!Series_summary\t\"Microarray data identified several genes with highly differential expression; e.g. Growth Differentiation Factor 15 (GDF-15), Activating Transcription Factor 3 (ATF3) and Amphiregulin (AREG). Pathway analysis revealed strong associations with ‘p53’ and ‘Diabetes Mellitus’ gene sets. Based on known function, we measured GDF-15 protein levels in 41 testicular patients during clinical follow-up. Pre-chemotherapy GDF-15 levels equalled controls. Throughout chemotherapy GDF-15, vWF and hsCRP levels increased, and were correlated at different time-points.\"\n",
      "!Series_summary\t\"An unbiased approach in a preclinical model revealed genes related to chemotherapy-induced endothelial damage, like GDF-15. The increases in plasma GDF-15 levels in testicular cancer patients during chemotherapy and its association with vWF and hsCRP suggest that GDF-15 is a potentially useful biomarker related to endothelial damage.\"\n",
      "!Series_overall_design\t\"In an acute-exposure setting, HMEC-1 were left untreated as controls or were treated with 0.3 (IC50 (concentration inhibiting cell survival by 50%)) or 1.5 ug/mL (IC90) bleomycin and 2.6 (IC50) and 12.9 uM (IC90) cisplatin for 6, 24 and 48 hours. In a chronic-exposure setting, lower doses were administrated (IC10; bleomycin 0.06 ug/mL or cisplatin 0.52 uM) two times a week; cells were collected for analysis at day 30. Administration of cisplatin had to be withheld at the 7th administration because of considerable cell dead, but was continued at full dose thereafter. Bleomycin could be administrated without disruption. Total RNA was isolated from HMEC-1 and pooled for each time point and drug concentration from 2 independent experiments.\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1528220</th>\n",
       "      <th>GSM1528221</th>\n",
       "      <th>GSM1528222</th>\n",
       "      <th>GSM1528223</th>\n",
       "      <th>GSM1528224</th>\n",
       "      <th>GSM1528225</th>\n",
       "      <th>GSM1528226</th>\n",
       "      <th>GSM1528227</th>\n",
       "      <th>GSM1528228</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1528314</th>\n",
       "      <th>GSM1528315</th>\n",
       "      <th>GSM1528316</th>\n",
       "      <th>GSM1528317</th>\n",
       "      <th>GSM1528318</th>\n",
       "      <th>GSM1528319</th>\n",
       "      <th>GSM1528320</th>\n",
       "      <th>GSM1528321</th>\n",
       "      <th>GSM1528322</th>\n",
       "      <th>GSM1528323</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>...</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "      <td>cell line: HMEC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "      <td>cell type: human microvascular endothelial cel...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "\n",
       "                                          GSM1528220  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528221  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528222  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528223  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528224  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528225  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528226  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528227  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528228  ...  \\\n",
       "0                                  cell line: HMEC-1  ...   \n",
       "1  cell type: human microvascular endothelial cel...  ...   \n",
       "\n",
       "                                          GSM1528314  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528315  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528316  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528317  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528318  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528319  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528320  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528321  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528322  \\\n",
       "0                                  cell line: HMEC-1   \n",
       "1  cell type: human microvascular endothelial cel...   \n",
       "\n",
       "                                          GSM1528323  \n",
       "0                                  cell line: HMEC-1  \n",
       "1  cell type: human microvascular endothelial cel...  \n",
       "\n",
       "[2 rows x 105 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM319520</th>\n",
       "      <th>GSM319521</th>\n",
       "      <th>GSM319522</th>\n",
       "      <th>GSM319523</th>\n",
       "      <th>GSM319524</th>\n",
       "      <th>GSM319525</th>\n",
       "      <th>GSM319526</th>\n",
       "      <th>GSM319527</th>\n",
       "      <th>GSM319528</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM319599</th>\n",
       "      <th>GSM319600</th>\n",
       "      <th>GSM319601</th>\n",
       "      <th>GSM319602</th>\n",
       "      <th>GSM319603</th>\n",
       "      <th>GSM319604</th>\n",
       "      <th>GSM319605</th>\n",
       "      <th>GSM319606</th>\n",
       "      <th>GSM319607</th>\n",
       "      <th>GSM319608</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Bladder</td>\n",
       "      <td>Anatomical sites: Bladder</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Bladder</td>\n",
       "      <td>Anatomical sites: Bladder</td>\n",
       "      <td>Anatomical sites: Prostate</td>\n",
       "      <td>Anatomical sites: Bladder</td>\n",
       "      <td>...</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Pancreas</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Stomach</td>\n",
       "      <td>Anatomical sites: Retroperiteum</td>\n",
       "      <td>Anatomical sites: Pancreas</td>\n",
       "      <td>Anatomical sites: Lymph node</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "      <td>Anatomical sites: Liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>...</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "      <td>gender: F</td>\n",
       "      <td>gender: M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>age: 65</td>\n",
       "      <td>age: 65</td>\n",
       "      <td>age: 65</td>\n",
       "      <td>age: 55</td>\n",
       "      <td>age: 55</td>\n",
       "      <td>age: 55</td>\n",
       "      <td>age: 45</td>\n",
       "      <td>age: 65</td>\n",
       "      <td>age: 65</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 96</td>\n",
       "      <td>age: 49</td>\n",
       "      <td>age: 66</td>\n",
       "      <td>age: 88</td>\n",
       "      <td>age: 70</td>\n",
       "      <td>age: 78</td>\n",
       "      <td>age: 67</td>\n",
       "      <td>age: 70</td>\n",
       "      <td>age: 68</td>\n",
       "      <td>age: 64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>TNM: Gleason 7</td>\n",
       "      <td>TNM: G4</td>\n",
       "      <td>...</td>\n",
       "      <td>TNM: 3</td>\n",
       "      <td>TNM: 4</td>\n",
       "      <td>TNM: 3</td>\n",
       "      <td>TNM: 3</td>\n",
       "      <td>TNM: 4</td>\n",
       "      <td>TNM: 2</td>\n",
       "      <td>TNM: 4</td>\n",
       "      <td>TNM: 3</td>\n",
       "      <td>TNM: 3</td>\n",
       "      <td>TNM: 3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession                GSM319520  \\\n",
       "0  !Sample_characteristics_ch1  Anatomical sites: Liver   \n",
       "1  !Sample_characteristics_ch1                gender: F   \n",
       "2  !Sample_characteristics_ch1                  age: 65   \n",
       "3  !Sample_characteristics_ch1                  TNM: G4   \n",
       "\n",
       "                   GSM319521                  GSM319522  \\\n",
       "0  Anatomical sites: Bladder  Anatomical sites: Bladder   \n",
       "1                  gender: M                  gender: M   \n",
       "2                    age: 65                    age: 65   \n",
       "3                    TNM: G4                    TNM: G4   \n",
       "\n",
       "                 GSM319523                GSM319524  \\\n",
       "0  Anatomical sites: Liver  Anatomical sites: Liver   \n",
       "1                gender: M                gender: F   \n",
       "2                  age: 55                  age: 55   \n",
       "3                  TNM: G4                  TNM: G4   \n",
       "\n",
       "                   GSM319525                  GSM319526  \\\n",
       "0  Anatomical sites: Bladder  Anatomical sites: Bladder   \n",
       "1                  gender: M                  gender: M   \n",
       "2                    age: 55                    age: 45   \n",
       "3                    TNM: G4                    TNM: G4   \n",
       "\n",
       "                    GSM319527                  GSM319528  ...  \\\n",
       "0  Anatomical sites: Prostate  Anatomical sites: Bladder  ...   \n",
       "1                   gender: M                  gender: M  ...   \n",
       "2                     age: 65                    age: 65  ...   \n",
       "3              TNM: Gleason 7                    TNM: G4  ...   \n",
       "\n",
       "                 GSM319599                   GSM319600  \\\n",
       "0  Anatomical sites: Liver  Anatomical sites: Pancreas   \n",
       "1                gender: F                   gender: F   \n",
       "2                  age: 96                     age: 49   \n",
       "3                   TNM: 3                      TNM: 4   \n",
       "\n",
       "                 GSM319601                GSM319602  \\\n",
       "0  Anatomical sites: Liver  Anatomical sites: Liver   \n",
       "1                gender: M                gender: M   \n",
       "2                  age: 66                  age: 88   \n",
       "3                   TNM: 3                   TNM: 3   \n",
       "\n",
       "                   GSM319603                        GSM319604  \\\n",
       "0  Anatomical sites: Stomach  Anatomical sites: Retroperiteum   \n",
       "1                  gender: M                        gender: M   \n",
       "2                    age: 70                          age: 78   \n",
       "3                     TNM: 4                           TNM: 2   \n",
       "\n",
       "                    GSM319605                     GSM319606  \\\n",
       "0  Anatomical sites: Pancreas  Anatomical sites: Lymph node   \n",
       "1                   gender: F                     gender: M   \n",
       "2                     age: 67                       age: 70   \n",
       "3                      TNM: 4                        TNM: 3   \n",
       "\n",
       "                 GSM319607                GSM319608  \n",
       "0  Anatomical sites: Liver  Anatomical sites: Liver  \n",
       "1                gender: F                gender: M  \n",
       "2                  age: 68                  age: 64  \n",
       "3                   TNM: 3                   TNM: 3  \n",
       "\n",
       "[4 rows x 90 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def parse_age(age_str):\n",
    "    if (age_str == \"age: -\"):\n",
    "        return age_str\n",
    "    elif '-' in age_str:\n",
    "        age_range = age_str.split('age: ')[1]\n",
    "        start_age, end_age = map(int, age_range.split('-'))\n",
    "        median_age = int(np.ceil((start_age + end_age)/2))\n",
    "        return f'age: {median_age}'\n",
    "    else:\n",
    "        return age_str  \n",
    "    \n",
    "clinical_data.iloc[2] = clinical_data.iloc[2].apply(parse_age)\n",
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'cell type: human microvascular endothelial cell line'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    if (trait == 'sample type: serum from patient with yolk sac tumor' or 'sample type: serum from patient with teratoma' or 'sample type: serum from patient with embryonal carcinoma' or\n",
    "        'sample type: serum from patient with embryonal carcinoma + yolk sac tumor' or 'sample type: serum from patient with embryonal carcinoma + teratoma' or\n",
    "        'sample type: serum from patient with seminoma'):\n",
    "        return 1  \n",
    "    else:\n",
    "        return 0  \n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string.lower() == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning:\n",
      "\n",
      "DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1438703</th>\n",
       "      <th>GSM1438704</th>\n",
       "      <th>GSM1438705</th>\n",
       "      <th>GSM1438706</th>\n",
       "      <th>GSM1438707</th>\n",
       "      <th>GSM1438708</th>\n",
       "      <th>GSM1438709</th>\n",
       "      <th>GSM1438710</th>\n",
       "      <th>GSM1438711</th>\n",
       "      <th>GSM1438712</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1438728</th>\n",
       "      <th>GSM1438729</th>\n",
       "      <th>GSM1438730</th>\n",
       "      <th>GSM1438731</th>\n",
       "      <th>GSM1438732</th>\n",
       "      <th>GSM1438733</th>\n",
       "      <th>GSM1438734</th>\n",
       "      <th>GSM1438735</th>\n",
       "      <th>GSM1438736</th>\n",
       "      <th>GSM1438737</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Testicular Cancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GSM1438703  GSM1438704  GSM1438705  GSM1438706  GSM1438707  \\\n",
       "Testicular Cancer           1           1           1           1           1   \n",
       "\n",
       "                   GSM1438708  GSM1438709  GSM1438710  GSM1438711  GSM1438712  \\\n",
       "Testicular Cancer           1           1           1           1           1   \n",
       "\n",
       "                   ...  GSM1438728  GSM1438729  GSM1438730  GSM1438731  \\\n",
       "Testicular Cancer  ...           1           1           1           1   \n",
       "\n",
       "                   GSM1438732  GSM1438733  GSM1438734  GSM1438735  GSM1438736  \\\n",
       "Testicular Cancer           1           1           1           1           1   \n",
       "\n",
       "                   GSM1438737  \n",
       "Testicular Cancer           1  \n",
       "\n",
       "[1 rows x 35 columns]"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1438703</th>\n",
       "      <th>GSM1438704</th>\n",
       "      <th>GSM1438705</th>\n",
       "      <th>GSM1438706</th>\n",
       "      <th>GSM1438707</th>\n",
       "      <th>GSM1438708</th>\n",
       "      <th>GSM1438709</th>\n",
       "      <th>GSM1438710</th>\n",
       "      <th>GSM1438711</th>\n",
       "      <th>GSM1438712</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1438728</th>\n",
       "      <th>GSM1438729</th>\n",
       "      <th>GSM1438730</th>\n",
       "      <th>GSM1438731</th>\n",
       "      <th>GSM1438732</th>\n",
       "      <th>GSM1438733</th>\n",
       "      <th>GSM1438734</th>\n",
       "      <th>GSM1438735</th>\n",
       "      <th>GSM1438736</th>\n",
       "      <th>GSM1438737</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a#-002307_B</th>\n",
       "      <td>5.842278</td>\n",
       "      <td>5.186990</td>\n",
       "      <td>-8.020862</td>\n",
       "      <td>4.256442</td>\n",
       "      <td>4.612815</td>\n",
       "      <td>5.217033</td>\n",
       "      <td>-0.603717</td>\n",
       "      <td>5.435251</td>\n",
       "      <td>3.867738</td>\n",
       "      <td>5.321154</td>\n",
       "      <td>...</td>\n",
       "      <td>5.441058</td>\n",
       "      <td>3.046009</td>\n",
       "      <td>6.003657</td>\n",
       "      <td>6.135078</td>\n",
       "      <td>5.370420</td>\n",
       "      <td>6.212634</td>\n",
       "      <td>5.985714</td>\n",
       "      <td>5.214142</td>\n",
       "      <td>4.578491</td>\n",
       "      <td>5.079505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-000377_A</th>\n",
       "      <td>-3.222858</td>\n",
       "      <td>-6.503963</td>\n",
       "      <td>-4.643490</td>\n",
       "      <td>0.949426</td>\n",
       "      <td>-8.546852</td>\n",
       "      <td>-6.698218</td>\n",
       "      <td>-6.865445</td>\n",
       "      <td>-5.403104</td>\n",
       "      <td>0.101707</td>\n",
       "      <td>-8.297680</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.240047</td>\n",
       "      <td>-8.201252</td>\n",
       "      <td>-7.662351</td>\n",
       "      <td>-8.063188</td>\n",
       "      <td>-4.468650</td>\n",
       "      <td>-7.213025</td>\n",
       "      <td>-6.777511</td>\n",
       "      <td>-5.475506</td>\n",
       "      <td>0.180072</td>\n",
       "      <td>0.408909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7b#-002404_B</th>\n",
       "      <td>-0.643702</td>\n",
       "      <td>5.445452</td>\n",
       "      <td>5.223694</td>\n",
       "      <td>4.519038</td>\n",
       "      <td>4.873827</td>\n",
       "      <td>5.475362</td>\n",
       "      <td>6.689619</td>\n",
       "      <td>-0.715029</td>\n",
       "      <td>3.764839</td>\n",
       "      <td>5.282783</td>\n",
       "      <td>...</td>\n",
       "      <td>5.487121</td>\n",
       "      <td>0.279509</td>\n",
       "      <td>6.047222</td>\n",
       "      <td>6.178059</td>\n",
       "      <td>-1.007834</td>\n",
       "      <td>6.255270</td>\n",
       "      <td>6.029359</td>\n",
       "      <td>3.958808</td>\n",
       "      <td>3.294935</td>\n",
       "      <td>3.818193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7b-002619_A</th>\n",
       "      <td>-8.538878</td>\n",
       "      <td>-12.678577</td>\n",
       "      <td>-8.127248</td>\n",
       "      <td>-6.813449</td>\n",
       "      <td>-10.991424</td>\n",
       "      <td>-11.159446</td>\n",
       "      <td>-11.013781</td>\n",
       "      <td>-10.235487</td>\n",
       "      <td>-10.001887</td>\n",
       "      <td>-10.271321</td>\n",
       "      <td>...</td>\n",
       "      <td>-11.753343</td>\n",
       "      <td>-10.540940</td>\n",
       "      <td>-11.435280</td>\n",
       "      <td>-11.226297</td>\n",
       "      <td>-8.359788</td>\n",
       "      <td>-11.168254</td>\n",
       "      <td>-9.711721</td>\n",
       "      <td>-11.853189</td>\n",
       "      <td>-10.120176</td>\n",
       "      <td>-5.492286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7c#-002405_B</th>\n",
       "      <td>5.866882</td>\n",
       "      <td>5.189973</td>\n",
       "      <td>4.959877</td>\n",
       "      <td>4.228723</td>\n",
       "      <td>4.596853</td>\n",
       "      <td>5.221008</td>\n",
       "      <td>6.480926</td>\n",
       "      <td>5.446426</td>\n",
       "      <td>5.114929</td>\n",
       "      <td>6.221021</td>\n",
       "      <td>...</td>\n",
       "      <td>5.199384</td>\n",
       "      <td>6.517703</td>\n",
       "      <td>5.780546</td>\n",
       "      <td>5.916303</td>\n",
       "      <td>5.126416</td>\n",
       "      <td>5.996417</td>\n",
       "      <td>5.762011</td>\n",
       "      <td>5.032583</td>\n",
       "      <td>4.548834</td>\n",
       "      <td>4.930120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      GSM1438703  GSM1438704  GSM1438705  GSM1438706  \\\n",
       "ID                                                                     \n",
       "hsa-let-7a#-002307_B    5.842278    5.186990   -8.020862    4.256442   \n",
       "hsa-let-7a-000377_A    -3.222858   -6.503963   -4.643490    0.949426   \n",
       "hsa-let-7b#-002404_B   -0.643702    5.445452    5.223694    4.519038   \n",
       "hsa-let-7b-002619_A    -8.538878  -12.678577   -8.127248   -6.813449   \n",
       "hsa-let-7c#-002405_B    5.866882    5.189973    4.959877    4.228723   \n",
       "\n",
       "                      GSM1438707  GSM1438708  GSM1438709  GSM1438710  \\\n",
       "ID                                                                     \n",
       "hsa-let-7a#-002307_B    4.612815    5.217033   -0.603717    5.435251   \n",
       "hsa-let-7a-000377_A    -8.546852   -6.698218   -6.865445   -5.403104   \n",
       "hsa-let-7b#-002404_B    4.873827    5.475362    6.689619   -0.715029   \n",
       "hsa-let-7b-002619_A   -10.991424  -11.159446  -11.013781  -10.235487   \n",
       "hsa-let-7c#-002405_B    4.596853    5.221008    6.480926    5.446426   \n",
       "\n",
       "                      GSM1438711  GSM1438712  ...  GSM1438728  GSM1438729  \\\n",
       "ID                                            ...                           \n",
       "hsa-let-7a#-002307_B    3.867738    5.321154  ...    5.441058    3.046009   \n",
       "hsa-let-7a-000377_A     0.101707   -8.297680  ...   -9.240047   -8.201252   \n",
       "hsa-let-7b#-002404_B    3.764839    5.282783  ...    5.487121    0.279509   \n",
       "hsa-let-7b-002619_A   -10.001887  -10.271321  ...  -11.753343  -10.540940   \n",
       "hsa-let-7c#-002405_B    5.114929    6.221021  ...    5.199384    6.517703   \n",
       "\n",
       "                      GSM1438730  GSM1438731  GSM1438732  GSM1438733  \\\n",
       "ID                                                                     \n",
       "hsa-let-7a#-002307_B    6.003657    6.135078    5.370420    6.212634   \n",
       "hsa-let-7a-000377_A    -7.662351   -8.063188   -4.468650   -7.213025   \n",
       "hsa-let-7b#-002404_B    6.047222    6.178059   -1.007834    6.255270   \n",
       "hsa-let-7b-002619_A   -11.435280  -11.226297   -8.359788  -11.168254   \n",
       "hsa-let-7c#-002405_B    5.780546    5.916303    5.126416    5.996417   \n",
       "\n",
       "                      GSM1438734  GSM1438735  GSM1438736  GSM1438737  \n",
       "ID                                                                    \n",
       "hsa-let-7a#-002307_B    5.985714    5.214142    4.578491    5.079505  \n",
       "hsa-let-7a-000377_A    -6.777511   -5.475506    0.180072    0.408909  \n",
       "hsa-let-7b#-002404_B    6.029359    3.958808    3.294935    3.818193  \n",
       "hsa-let-7b-002619_A    -9.711721  -11.853189  -10.120176   -5.492286  \n",
       "hsa-let-7c#-002405_B    5.762011    5.032583    4.548834    4.930120  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['hsa-let-7a#-002307_B', 'hsa-let-7a-000377_A', 'hsa-let-7b#-002404_B', 'hsa-let-7b-002619_A', 'hsa-let-7c#-002405_B'], 'miRNA_ID': ['hsa-let-7a', 'hsa-let-7a', 'hsa-let-7b', 'hsa-let-7b', 'hsa-let-7c']}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'miRNA_ID'], dtype='object')"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'Symbol'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM694425</th>\n",
       "      <th>GSM694426</th>\n",
       "      <th>GSM694427</th>\n",
       "      <th>GSM694428</th>\n",
       "      <th>GSM694429</th>\n",
       "      <th>GSM694430</th>\n",
       "      <th>GSM694431</th>\n",
       "      <th>GSM694432</th>\n",
       "      <th>GSM694433</th>\n",
       "      <th>GSM694434</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM696043</th>\n",
       "      <th>GSM696044</th>\n",
       "      <th>GSM696045</th>\n",
       "      <th>GSM696046</th>\n",
       "      <th>GSM696047</th>\n",
       "      <th>GSM696048</th>\n",
       "      <th>GSM696049</th>\n",
       "      <th>GSM696050</th>\n",
       "      <th>GSM696051</th>\n",
       "      <th>GSM696052</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AATK</th>\n",
       "      <td>0.763330</td>\n",
       "      <td>0.462200</td>\n",
       "      <td>0.642997</td>\n",
       "      <td>0.478480</td>\n",
       "      <td>0.662853</td>\n",
       "      <td>0.542033</td>\n",
       "      <td>0.634990</td>\n",
       "      <td>0.866760</td>\n",
       "      <td>0.907427</td>\n",
       "      <td>0.769250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.586590</td>\n",
       "      <td>0.868620</td>\n",
       "      <td>0.902693</td>\n",
       "      <td>0.082113</td>\n",
       "      <td>0.844367</td>\n",
       "      <td>0.215980</td>\n",
       "      <td>0.262597</td>\n",
       "      <td>0.217850</td>\n",
       "      <td>0.693370</td>\n",
       "      <td>0.873490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA1</th>\n",
       "      <td>0.055445</td>\n",
       "      <td>0.064085</td>\n",
       "      <td>0.030930</td>\n",
       "      <td>0.026370</td>\n",
       "      <td>0.052145</td>\n",
       "      <td>0.041280</td>\n",
       "      <td>0.040625</td>\n",
       "      <td>0.119855</td>\n",
       "      <td>0.031850</td>\n",
       "      <td>0.044510</td>\n",
       "      <td>...</td>\n",
       "      <td>0.486990</td>\n",
       "      <td>0.270730</td>\n",
       "      <td>0.091300</td>\n",
       "      <td>0.034815</td>\n",
       "      <td>0.035905</td>\n",
       "      <td>0.025070</td>\n",
       "      <td>0.061780</td>\n",
       "      <td>0.030165</td>\n",
       "      <td>0.036575</td>\n",
       "      <td>0.021745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB1</th>\n",
       "      <td>0.029260</td>\n",
       "      <td>0.033270</td>\n",
       "      <td>0.021860</td>\n",
       "      <td>0.289410</td>\n",
       "      <td>0.029440</td>\n",
       "      <td>0.030050</td>\n",
       "      <td>0.045110</td>\n",
       "      <td>0.631570</td>\n",
       "      <td>0.025270</td>\n",
       "      <td>0.431280</td>\n",
       "      <td>...</td>\n",
       "      <td>0.149060</td>\n",
       "      <td>0.112050</td>\n",
       "      <td>0.815110</td>\n",
       "      <td>0.023870</td>\n",
       "      <td>0.859140</td>\n",
       "      <td>0.070090</td>\n",
       "      <td>0.025860</td>\n",
       "      <td>0.264610</td>\n",
       "      <td>0.140650</td>\n",
       "      <td>0.019740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB4</th>\n",
       "      <td>0.917533</td>\n",
       "      <td>0.923730</td>\n",
       "      <td>0.820663</td>\n",
       "      <td>0.412440</td>\n",
       "      <td>0.831660</td>\n",
       "      <td>0.892687</td>\n",
       "      <td>0.892533</td>\n",
       "      <td>0.867287</td>\n",
       "      <td>0.913630</td>\n",
       "      <td>0.906713</td>\n",
       "      <td>...</td>\n",
       "      <td>0.535997</td>\n",
       "      <td>0.142330</td>\n",
       "      <td>0.468137</td>\n",
       "      <td>0.494043</td>\n",
       "      <td>0.430917</td>\n",
       "      <td>0.483337</td>\n",
       "      <td>0.537487</td>\n",
       "      <td>0.561760</td>\n",
       "      <td>0.781263</td>\n",
       "      <td>0.925660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCC2</th>\n",
       "      <td>0.644555</td>\n",
       "      <td>0.878110</td>\n",
       "      <td>0.798410</td>\n",
       "      <td>0.125310</td>\n",
       "      <td>0.631210</td>\n",
       "      <td>0.894670</td>\n",
       "      <td>0.709640</td>\n",
       "      <td>0.938555</td>\n",
       "      <td>0.947535</td>\n",
       "      <td>0.938880</td>\n",
       "      <td>...</td>\n",
       "      <td>0.899720</td>\n",
       "      <td>0.899600</td>\n",
       "      <td>0.897065</td>\n",
       "      <td>0.136490</td>\n",
       "      <td>0.566495</td>\n",
       "      <td>0.446390</td>\n",
       "      <td>0.628845</td>\n",
       "      <td>0.604705</td>\n",
       "      <td>0.806435</td>\n",
       "      <td>0.954325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZIM3</th>\n",
       "      <td>0.509543</td>\n",
       "      <td>0.622737</td>\n",
       "      <td>0.807350</td>\n",
       "      <td>0.426720</td>\n",
       "      <td>0.908350</td>\n",
       "      <td>0.734730</td>\n",
       "      <td>0.908807</td>\n",
       "      <td>0.666313</td>\n",
       "      <td>0.943337</td>\n",
       "      <td>0.920040</td>\n",
       "      <td>...</td>\n",
       "      <td>0.952477</td>\n",
       "      <td>0.112767</td>\n",
       "      <td>0.796520</td>\n",
       "      <td>0.499650</td>\n",
       "      <td>0.893730</td>\n",
       "      <td>0.711397</td>\n",
       "      <td>0.622020</td>\n",
       "      <td>0.714327</td>\n",
       "      <td>0.854890</td>\n",
       "      <td>0.953750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMYND10</th>\n",
       "      <td>0.043255</td>\n",
       "      <td>0.054505</td>\n",
       "      <td>0.042720</td>\n",
       "      <td>0.072095</td>\n",
       "      <td>0.072390</td>\n",
       "      <td>0.027790</td>\n",
       "      <td>0.323365</td>\n",
       "      <td>0.665285</td>\n",
       "      <td>0.077090</td>\n",
       "      <td>0.318725</td>\n",
       "      <td>...</td>\n",
       "      <td>0.920440</td>\n",
       "      <td>0.742160</td>\n",
       "      <td>0.418530</td>\n",
       "      <td>0.106025</td>\n",
       "      <td>0.041645</td>\n",
       "      <td>0.060450</td>\n",
       "      <td>0.062975</td>\n",
       "      <td>0.045760</td>\n",
       "      <td>0.068870</td>\n",
       "      <td>0.048235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF215</th>\n",
       "      <td>0.206340</td>\n",
       "      <td>0.115455</td>\n",
       "      <td>0.056785</td>\n",
       "      <td>0.036935</td>\n",
       "      <td>0.565590</td>\n",
       "      <td>0.085890</td>\n",
       "      <td>0.073600</td>\n",
       "      <td>0.078520</td>\n",
       "      <td>0.081040</td>\n",
       "      <td>0.130275</td>\n",
       "      <td>...</td>\n",
       "      <td>0.354250</td>\n",
       "      <td>0.167925</td>\n",
       "      <td>0.506625</td>\n",
       "      <td>0.083815</td>\n",
       "      <td>0.795100</td>\n",
       "      <td>0.075735</td>\n",
       "      <td>0.090965</td>\n",
       "      <td>0.291520</td>\n",
       "      <td>0.138770</td>\n",
       "      <td>0.148765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF264</th>\n",
       "      <td>0.248280</td>\n",
       "      <td>0.420710</td>\n",
       "      <td>0.193010</td>\n",
       "      <td>0.137505</td>\n",
       "      <td>0.663030</td>\n",
       "      <td>0.243080</td>\n",
       "      <td>0.186635</td>\n",
       "      <td>0.317830</td>\n",
       "      <td>0.351480</td>\n",
       "      <td>0.233920</td>\n",
       "      <td>...</td>\n",
       "      <td>0.445250</td>\n",
       "      <td>0.156665</td>\n",
       "      <td>0.282680</td>\n",
       "      <td>0.101420</td>\n",
       "      <td>0.366665</td>\n",
       "      <td>0.086490</td>\n",
       "      <td>0.120520</td>\n",
       "      <td>0.074880</td>\n",
       "      <td>0.227085</td>\n",
       "      <td>0.207445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP3</th>\n",
       "      <td>0.120985</td>\n",
       "      <td>0.142005</td>\n",
       "      <td>0.232720</td>\n",
       "      <td>0.020355</td>\n",
       "      <td>0.155695</td>\n",
       "      <td>0.152195</td>\n",
       "      <td>0.343785</td>\n",
       "      <td>0.287175</td>\n",
       "      <td>0.436445</td>\n",
       "      <td>0.468455</td>\n",
       "      <td>...</td>\n",
       "      <td>0.877695</td>\n",
       "      <td>0.148820</td>\n",
       "      <td>0.335325</td>\n",
       "      <td>0.161290</td>\n",
       "      <td>0.046090</td>\n",
       "      <td>0.045570</td>\n",
       "      <td>0.195115</td>\n",
       "      <td>0.085220</td>\n",
       "      <td>0.256345</td>\n",
       "      <td>0.379010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>761 rows × 1628 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GSM694425  GSM694426  GSM694427  GSM694428  GSM694429  GSM694430  \\\n",
       "AATK      0.763330   0.462200   0.642997   0.478480   0.662853   0.542033   \n",
       "ABCA1     0.055445   0.064085   0.030930   0.026370   0.052145   0.041280   \n",
       "ABCB1     0.029260   0.033270   0.021860   0.289410   0.029440   0.030050   \n",
       "ABCB4     0.917533   0.923730   0.820663   0.412440   0.831660   0.892687   \n",
       "ABCC2     0.644555   0.878110   0.798410   0.125310   0.631210   0.894670   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "ZIM3      0.509543   0.622737   0.807350   0.426720   0.908350   0.734730   \n",
       "ZMYND10   0.043255   0.054505   0.042720   0.072095   0.072390   0.027790   \n",
       "ZNF215    0.206340   0.115455   0.056785   0.036935   0.565590   0.085890   \n",
       "ZNF264    0.248280   0.420710   0.193010   0.137505   0.663030   0.243080   \n",
       "ZP3       0.120985   0.142005   0.232720   0.020355   0.155695   0.152195   \n",
       "\n",
       "         GSM694431  GSM694432  GSM694433  GSM694434  ...  GSM696043  \\\n",
       "AATK      0.634990   0.866760   0.907427   0.769250  ...   0.586590   \n",
       "ABCA1     0.040625   0.119855   0.031850   0.044510  ...   0.486990   \n",
       "ABCB1     0.045110   0.631570   0.025270   0.431280  ...   0.149060   \n",
       "ABCB4     0.892533   0.867287   0.913630   0.906713  ...   0.535997   \n",
       "ABCC2     0.709640   0.938555   0.947535   0.938880  ...   0.899720   \n",
       "...            ...        ...        ...        ...  ...        ...   \n",
       "ZIM3      0.908807   0.666313   0.943337   0.920040  ...   0.952477   \n",
       "ZMYND10   0.323365   0.665285   0.077090   0.318725  ...   0.920440   \n",
       "ZNF215    0.073600   0.078520   0.081040   0.130275  ...   0.354250   \n",
       "ZNF264    0.186635   0.317830   0.351480   0.233920  ...   0.445250   \n",
       "ZP3       0.343785   0.287175   0.436445   0.468455  ...   0.877695   \n",
       "\n",
       "         GSM696044  GSM696045  GSM696046  GSM696047  GSM696048  GSM696049  \\\n",
       "AATK      0.868620   0.902693   0.082113   0.844367   0.215980   0.262597   \n",
       "ABCA1     0.270730   0.091300   0.034815   0.035905   0.025070   0.061780   \n",
       "ABCB1     0.112050   0.815110   0.023870   0.859140   0.070090   0.025860   \n",
       "ABCB4     0.142330   0.468137   0.494043   0.430917   0.483337   0.537487   \n",
       "ABCC2     0.899600   0.897065   0.136490   0.566495   0.446390   0.628845   \n",
       "...            ...        ...        ...        ...        ...        ...   \n",
       "ZIM3      0.112767   0.796520   0.499650   0.893730   0.711397   0.622020   \n",
       "ZMYND10   0.742160   0.418530   0.106025   0.041645   0.060450   0.062975   \n",
       "ZNF215    0.167925   0.506625   0.083815   0.795100   0.075735   0.090965   \n",
       "ZNF264    0.156665   0.282680   0.101420   0.366665   0.086490   0.120520   \n",
       "ZP3       0.148820   0.335325   0.161290   0.046090   0.045570   0.195115   \n",
       "\n",
       "         GSM696050  GSM696051  GSM696052  \n",
       "AATK      0.217850   0.693370   0.873490  \n",
       "ABCA1     0.030165   0.036575   0.021745  \n",
       "ABCB1     0.264610   0.140650   0.019740  \n",
       "ABCB4     0.561760   0.781263   0.925660  \n",
       "ABCC2     0.604705   0.806435   0.954325  \n",
       "...            ...        ...        ...  \n",
       "ZIM3      0.714327   0.854890   0.953750  \n",
       "ZMYND10   0.045760   0.068870   0.048235  \n",
       "ZNF215    0.291520   0.138770   0.148765  \n",
       "ZNF264    0.074880   0.227085   0.207445  \n",
       "ZP3       0.085220   0.256345   0.379010  \n",
       "\n",
       "[761 rows x 1628 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Testicular Cancer</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN2</th>\n",
       "      <th>ZSWIM1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM319520</th>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.822356</td>\n",
       "      <td>-0.786868</td>\n",
       "      <td>-0.237055</td>\n",
       "      <td>0.167821</td>\n",
       "      <td>0.261164</td>\n",
       "      <td>1.618413</td>\n",
       "      <td>-0.518816</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.15557</td>\n",
       "      <td>0.260321</td>\n",
       "      <td>0.567675</td>\n",
       "      <td>0.523961</td>\n",
       "      <td>0.949124</td>\n",
       "      <td>-0.232081</td>\n",
       "      <td>0.569597</td>\n",
       "      <td>0.576858</td>\n",
       "      <td>0.135833</td>\n",
       "      <td>0.584064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319521</th>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.559114</td>\n",
       "      <td>-0.651795</td>\n",
       "      <td>-0.846372</td>\n",
       "      <td>0.193443</td>\n",
       "      <td>0.539628</td>\n",
       "      <td>0.135394</td>\n",
       "      <td>-0.286925</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.15805</td>\n",
       "      <td>-0.520908</td>\n",
       "      <td>0.646224</td>\n",
       "      <td>0.620909</td>\n",
       "      <td>1.084455</td>\n",
       "      <td>-0.106472</td>\n",
       "      <td>0.037375</td>\n",
       "      <td>0.55263</td>\n",
       "      <td>-0.182003</td>\n",
       "      <td>0.805389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319522</th>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.659479</td>\n",
       "      <td>-0.020965</td>\n",
       "      <td>-0.258145</td>\n",
       "      <td>0.050393</td>\n",
       "      <td>0.588356</td>\n",
       "      <td>0.349506</td>\n",
       "      <td>-0.425712</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.04737</td>\n",
       "      <td>-0.236573</td>\n",
       "      <td>0.636459</td>\n",
       "      <td>0.61771</td>\n",
       "      <td>0.974492</td>\n",
       "      <td>-0.386168</td>\n",
       "      <td>0.259272</td>\n",
       "      <td>0.733197</td>\n",
       "      <td>0.151216</td>\n",
       "      <td>0.807542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319523</th>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.259253</td>\n",
       "      <td>-0.234597</td>\n",
       "      <td>-0.034119</td>\n",
       "      <td>0.166468</td>\n",
       "      <td>0.176863</td>\n",
       "      <td>1.803988</td>\n",
       "      <td>-0.443632</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.126293</td>\n",
       "      <td>0.063562</td>\n",
       "      <td>0.530385</td>\n",
       "      <td>0.213963</td>\n",
       "      <td>0.570189</td>\n",
       "      <td>-0.194726</td>\n",
       "      <td>0.361144</td>\n",
       "      <td>0.289863</td>\n",
       "      <td>-0.553081</td>\n",
       "      <td>0.790123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319524</th>\n",
       "      <td>0.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.943809</td>\n",
       "      <td>-0.567291</td>\n",
       "      <td>-0.128662</td>\n",
       "      <td>0.278019</td>\n",
       "      <td>0.511934</td>\n",
       "      <td>1.529365</td>\n",
       "      <td>-0.453456</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.55456</td>\n",
       "      <td>0.060705</td>\n",
       "      <td>0.630919</td>\n",
       "      <td>0.41537</td>\n",
       "      <td>0.951013</td>\n",
       "      <td>-0.222035</td>\n",
       "      <td>0.375667</td>\n",
       "      <td>0.558456</td>\n",
       "      <td>0.002504</td>\n",
       "      <td>0.893576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319604</th>\n",
       "      <td>0.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.278891</td>\n",
       "      <td>-0.486858</td>\n",
       "      <td>-0.298108</td>\n",
       "      <td>0.195726</td>\n",
       "      <td>0.634411</td>\n",
       "      <td>0.851169</td>\n",
       "      <td>-0.426483</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.104783</td>\n",
       "      <td>0.082673</td>\n",
       "      <td>0.534862</td>\n",
       "      <td>0.184389</td>\n",
       "      <td>0.641579</td>\n",
       "      <td>-0.429776</td>\n",
       "      <td>0.41384</td>\n",
       "      <td>0.574886</td>\n",
       "      <td>-0.067703</td>\n",
       "      <td>0.651082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.409083</td>\n",
       "      <td>-0.811472</td>\n",
       "      <td>-0.237879</td>\n",
       "      <td>0.393832</td>\n",
       "      <td>0.700658</td>\n",
       "      <td>0.919927</td>\n",
       "      <td>-0.224048</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.982962</td>\n",
       "      <td>-0.138692</td>\n",
       "      <td>0.684017</td>\n",
       "      <td>0.205511</td>\n",
       "      <td>0.776389</td>\n",
       "      <td>-0.326719</td>\n",
       "      <td>0.35569</td>\n",
       "      <td>0.581518</td>\n",
       "      <td>-0.041377</td>\n",
       "      <td>0.700126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319606</th>\n",
       "      <td>0.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.147403</td>\n",
       "      <td>-1.046497</td>\n",
       "      <td>-0.086671</td>\n",
       "      <td>0.334379</td>\n",
       "      <td>0.336263</td>\n",
       "      <td>0.107917</td>\n",
       "      <td>-0.17182</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.911489</td>\n",
       "      <td>0.096947</td>\n",
       "      <td>0.572695</td>\n",
       "      <td>0.156804</td>\n",
       "      <td>0.169242</td>\n",
       "      <td>-1.046497</td>\n",
       "      <td>0.429288</td>\n",
       "      <td>0.363679</td>\n",
       "      <td>0.078083</td>\n",
       "      <td>0.631074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319607</th>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.726922</td>\n",
       "      <td>-0.645779</td>\n",
       "      <td>-0.063774</td>\n",
       "      <td>0.241135</td>\n",
       "      <td>0.455403</td>\n",
       "      <td>0.961991</td>\n",
       "      <td>-0.231324</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.014851</td>\n",
       "      <td>0.039325</td>\n",
       "      <td>0.664898</td>\n",
       "      <td>0.723621</td>\n",
       "      <td>1.166881</td>\n",
       "      <td>-0.86736</td>\n",
       "      <td>0.691084</td>\n",
       "      <td>0.537276</td>\n",
       "      <td>0.095422</td>\n",
       "      <td>0.715506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM319608</th>\n",
       "      <td>0.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.661675</td>\n",
       "      <td>-0.858728</td>\n",
       "      <td>-0.231011</td>\n",
       "      <td>0.315927</td>\n",
       "      <td>0.792733</td>\n",
       "      <td>0.53547</td>\n",
       "      <td>-0.209366</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.090714</td>\n",
       "      <td>-0.125369</td>\n",
       "      <td>0.758507</td>\n",
       "      <td>0.416704</td>\n",
       "      <td>1.000228</td>\n",
       "      <td>-0.304816</td>\n",
       "      <td>0.1291</td>\n",
       "      <td>0.77109</td>\n",
       "      <td>-0.357784</td>\n",
       "      <td>0.704644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 10560 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Testicular Cancer   Age Gender       A2M    A4GALT     A4GNT  \\\n",
       "GSM319520               0.0  65.0    1.0  1.822356 -0.786868 -0.237055   \n",
       "GSM319521               0.0  65.0    0.0  1.559114 -0.651795 -0.846372   \n",
       "GSM319522               0.0  65.0    0.0  1.659479 -0.020965 -0.258145   \n",
       "GSM319523               0.0  55.0    0.0  2.259253 -0.234597 -0.034119   \n",
       "GSM319524               0.0  55.0    1.0  1.943809 -0.567291 -0.128662   \n",
       "...                     ...   ...    ...       ...       ...       ...   \n",
       "GSM319604               0.0  78.0    0.0  1.278891 -0.486858 -0.298108   \n",
       "GSM319605               0.0  67.0    1.0  1.409083 -0.811472 -0.237879   \n",
       "GSM319606               0.0  70.0    0.0  1.147403 -1.046497 -0.086671   \n",
       "GSM319607               0.0  68.0    1.0  1.726922 -0.645779 -0.063774   \n",
       "GSM319608               0.0  64.0    0.0  1.661675 -0.858728 -0.231011   \n",
       "\n",
       "               AAAS      AACS     AADAC      AAK1  ...    ZSCAN2    ZSWIM1  \\\n",
       "GSM319520  0.167821  0.261164  1.618413 -0.518816  ...  -0.15557  0.260321   \n",
       "GSM319521  0.193443  0.539628  0.135394 -0.286925  ...  -0.15805 -0.520908   \n",
       "GSM319522  0.050393  0.588356  0.349506 -0.425712  ...  -1.04737 -0.236573   \n",
       "GSM319523  0.166468  0.176863  1.803988 -0.443632  ... -0.126293  0.063562   \n",
       "GSM319524  0.278019  0.511934  1.529365 -0.453456  ...  -0.55456  0.060705   \n",
       "...             ...       ...       ...       ...  ...       ...       ...   \n",
       "GSM319604  0.195726  0.634411  0.851169 -0.426483  ... -1.104783  0.082673   \n",
       "GSM319605  0.393832  0.700658  0.919927 -0.224048  ... -0.982962 -0.138692   \n",
       "GSM319606  0.334379  0.336263  0.107917  -0.17182  ... -0.911489  0.096947   \n",
       "GSM319607  0.241135  0.455403  0.961991 -0.231324  ... -1.014851  0.039325   \n",
       "GSM319608  0.315927  0.792733   0.53547 -0.209366  ... -0.090714 -0.125369   \n",
       "\n",
       "               ZW10    ZWILCH     ZWINT      ZXDB      ZXDC       ZYX  \\\n",
       "GSM319520  0.567675  0.523961  0.949124 -0.232081  0.569597  0.576858   \n",
       "GSM319521  0.646224  0.620909  1.084455 -0.106472  0.037375   0.55263   \n",
       "GSM319522  0.636459   0.61771  0.974492 -0.386168  0.259272  0.733197   \n",
       "GSM319523  0.530385  0.213963  0.570189 -0.194726  0.361144  0.289863   \n",
       "GSM319524  0.630919   0.41537  0.951013 -0.222035  0.375667  0.558456   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "GSM319604  0.534862  0.184389  0.641579 -0.429776   0.41384  0.574886   \n",
       "GSM319605  0.684017  0.205511  0.776389 -0.326719   0.35569  0.581518   \n",
       "GSM319606  0.572695  0.156804  0.169242 -1.046497  0.429288  0.363679   \n",
       "GSM319607  0.664898  0.723621  1.166881  -0.86736  0.691084  0.537276   \n",
       "GSM319608  0.758507  0.416704  1.000228 -0.304816    0.1291   0.77109   \n",
       "\n",
       "              ZZEF1      ZZZ3  \n",
       "GSM319520  0.135833  0.584064  \n",
       "GSM319521 -0.182003  0.805389  \n",
       "GSM319522  0.151216  0.807542  \n",
       "GSM319523 -0.553081  0.790123  \n",
       "GSM319524  0.002504  0.893576  \n",
       "...             ...       ...  \n",
       "GSM319604 -0.067703  0.651082  \n",
       "GSM319605 -0.041377  0.700126  \n",
       "GSM319606  0.078083  0.631074  \n",
       "GSM319607  0.095422  0.715506  \n",
       "GSM319608 -0.357784  0.704644  \n",
       "\n",
       "[88 rows x 10560 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 88 samples.\n",
      "For the feature 'Testicular Cancer', the least common label is '1.0' with 10 occurrences. This represents 11.36% of the dataset.\n",
      "The distribution of the feature 'Testicular Cancer' in this dataset is fine.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 44.75\n",
      "  50% (Median): 55.0\n",
      "  75%: 65.25\n",
      "Min: 15.0\n",
      "Max: 96.0\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '1.0' with 35 occurrences. This represents 39.77% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
