{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 1. Basic setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:19.918139380Z",
     "start_time": "2024-02-27T00:22:19.235047906Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "utils.py has been loaded\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "sys.path.append('..')\n",
    "from utils import *\n",
    "\n",
    "USER = \"Jiayi\"\n",
    "DATA_ROOT = '/Users/legion/Desktop/Courses/IS389/data'   \n",
    "OUTPUT_ROOT = '/Users/legion/Desktop/Courses/IS389/output2'\n",
    "TRAIT = 'Uterine Carcinosarcoma'\n",
    "\n",
    "OUTPUT_DIR = os.path.join(OUTPUT_ROOT, USER, '-'.join(TRAIT.split()))\n",
    "JSON_PATH = os.path.join(OUTPUT_DIR, \"cohort_info.json\")\n",
    "if not os.path.exists(OUTPUT_DIR):\n",
    "    os.makedirs(OUTPUT_DIR, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# 2. Data preprocessing and selection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2.1. The TCGA Xena dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In TCGA Xena, there is either zero or one cohort related to the trait. We search the names of subdirectories to see if any matches the trait. If a match is found, we directly obtain the file paths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:21.140790869Z",
     "start_time": "2024-02-27T00:22:21.133208795Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA_Adrenocortical_Cancer_(ACC)',\n",
       " 'TCGA_Breast_Cancer_(BRCA)',\n",
       " 'TCGA_Cervical_Cancer_(CESC)',\n",
       " 'TCGA_Kidney_Chromophobe_(KICH)',\n",
       " 'TCGA_Kidney_Papillary_Cell_Carcinoma_(KIRP)',\n",
       " 'TCGA_Lower_Grade_Glioma_(LGG)',\n",
       " 'TCGA_Melanoma_(SKCM)',\n",
       " 'TCGA_Mesothelioma_(MESO)',\n",
       " 'TCGA_Testicular_Cancer_(TGCT)',\n",
       " 'TCGA_Uterine_Carcinosarcoma_(UCS)']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'TCGA'\n",
    "dataset_dir = os.path.join(DATA_ROOT, dataset)\n",
    "os.listdir(dataset_dir)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:21.985144212Z",
     "start_time": "2024-02-27T00:22:21.981084968Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\TCGA\\\\TCGA_Uterine_Carcinosarcoma_(UCS)\\\\TCGA.UCS.sampleMap_UCS_clinicalMatrix',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\TCGA\\\\TCGA_Uterine_Carcinosarcoma_(UCS)\\\\TCGA.UCS.sampleMap_HiSeqV2_PANCAN.gz')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_subdir = \"TCGA_Uterine_Carcinosarcoma_(UCS)\"\n",
    "cohort = 'Xena'\n",
    "trait_type = 'binary'\n",
    "is_available = True\n",
    "\n",
    "cohort_dir = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "clinical_data_file, genetic_data_file = xena_get_relevant_filepaths(cohort_dir)\n",
    "clinical_data_file, genetic_data_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.794819905Z",
     "start_time": "2024-02-27T00:22:22.357077391Z"
    },
    "id": "MudwB-_iz7sc"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "clinical_data = pd.read_csv(clinical_data_file, sep='\\t', index_col=0)\n",
    "genetic_data = pd.read_csv(genetic_data_file, compression='gzip', sep='\\t', index_col=0)\n",
    "age_col = gender_col = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.798785383Z",
     "start_time": "2024-02-27T00:22:25.796474039Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CDE_ID_3226963',\n",
       " '_INTEGRATION',\n",
       " '_PATIENT',\n",
       " '_cohort',\n",
       " '_primary_disease',\n",
       " '_primary_site',\n",
       " 'additional_pharmaceutical_therapy',\n",
       " 'additional_radiation_therapy',\n",
       " 'age_at_initial_pathologic_diagnosis',\n",
       " 'aln_pos_ihc']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, clinical_data_cols = check_rows_and_columns(clinical_data)\n",
    "clinical_data_cols[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T20:20:11.841042813Z",
     "start_time": "2023-12-28T20:20:11.834458206Z"
    },
    "collapsed": false
   },
   "source": [
    "Read all the column names in the clinical dataset, to find the columns that record information about age or gender.\n",
    "Reference prompt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.802838875Z",
     "start_time": "2024-02-27T00:22:25.798659327Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nBelow is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\\ncandidate_age_cols = [col_name1, col_name2, ...]\\ncandidate_gender_cols = [col_name1, col_name2, ...]\\nIf no columns match a criterion, please provide an empty list.\\n\\nColumn names:\\n['CDE_ID_3226963', '_INTEGRATION', '_PATIENT', '_cohort', '_primary_disease', '_primary_site', 'additional_pharmaceutical_therapy', 'additional_radiation_therapy', 'age_at_initial_pathologic_diagnosis', 'aln_pos_ihc', 'aln_pos_light_micro', 'anatomic_neoplasm_subdivision', 'bcr_followup_barcode', 'bcr_patient_barcode', 'bcr_sample_barcode', 'birth_control_pill_history_usage_category', 'clinical_stage', 'colorectal_cancer', 'days_to_birth', 'days_to_collection', 'days_to_death', 'days_to_initial_pathologic_diagnosis', 'days_to_last_followup', 'days_to_new_tumor_event_additional_surgery_procedure', 'days_to_new_tumor_event_after_initial_treatment', 'diabetes', 'form_completion_date', 'gender', 'height', 'histological_type', 'history_of_neoadjuvant_treatment', 'horm_ther', 'hypertension', 'icd_10', 'icd_o_3_histology', 'icd_o_3_site', 'informed_consent_verified', 'initial_pathologic_diagnosis_method', 'initial_weight', 'is_ffpe', 'lost_follow_up', 'menopause_status', 'new_neoplasm_event_occurrence_anatomic_site', 'new_neoplasm_event_type', 'new_neoplasm_occurrence_anatomic_site_text', 'new_tumor_event_additional_surgery_procedure', 'new_tumor_event_after_initial_treatment', 'oct_embedded', 'other_dx', 'pathology_report_file_name', 'patient_id', 'pct_tumor_invasion', 'peritoneal_wash', 'person_neoplasm_cancer_status', 'pln_pos_ihc', 'pln_pos_light_micro', 'postoperative_rx_tx', 'pregnancies', 'primary_therapy_outcome_success', 'prior_tamoxifen_administered_usage_category', 'radiation_therapy', 'recurrence_second_surgery_neoplasm_surgical_procedure_name', 'residual_tumor', 'sample_type', 'sample_type_id', 'surgical_approach', 'system_version', 'tissue_prospective_collection_indicator', 'tissue_retrospective_collection_indicator', 'tissue_source_site', 'total_aor_lnp', 'total_aor_lnr', 'total_pelv_lnp', 'total_pelv_lnr', 'tumor_tissue_site', 'vial_number', 'vital_status', 'weight', 'year_of_initial_pathologic_diagnosis', '_GENOMIC_ID_TCGA_UCS_miRNA_HiSeq', '_GENOMIC_ID_TCGA_UCS_RPPA', '_GENOMIC_ID_data/public/TCGA/UCS/miRNA_HiSeq_gene', '_GENOMIC_ID_TCGA_UCS_exp_HiSeqV2_exon', '_GENOMIC_ID_TCGA_UCS_exp_HiSeqV2', '_GENOMIC_ID_TCGA_UCS_mutation_bcm_gene', '_GENOMIC_ID_TCGA_UCS_exp_HiSeqV2_percentile', '_GENOMIC_ID_TCGA_UCS_mutation_curated_broad_gene', '_GENOMIC_ID_TCGA_UCS_gistic2thd', '_GENOMIC_ID_TCGA_UCS_exp_HiSeqV2_PANCAN', '_GENOMIC_ID_TCGA_UCS_mutation_broad_gene', '_GENOMIC_ID_TCGA_UCS_PDMRNAseqCNV', '_GENOMIC_ID_TCGA_UCS_gistic2', '_GENOMIC_ID_TCGA_UCS_mutation_bcgsc_gene', '_GENOMIC_ID_TCGA_UCS_hMethyl450', '_GENOMIC_ID_TCGA_UCS_PDMRNAseq']\\n\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f'''\n",
    "Below is a list of column names from a biomedical dataset. Please examine it and identify the columns that are likely to contain information about patients' age. Additionally, please do the same for columns that may hold data on patients' gender. Please provide your answer by strictly following this format, without redundant words:\n",
    "candidate_age_cols = [col_name1, col_name2, ...]\n",
    "candidate_gender_cols = [col_name1, col_name2, ...]\n",
    "If no columns match a criterion, please provide an empty list.\n",
    "\n",
    "Column names:\n",
    "{clinical_data_cols}\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:25.864092575Z",
     "start_time": "2024-02-27T00:22:25.802001989Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "candidate_age_cols = ['age_at_initial_pathologic_diagnosis']\n",
    "candidate_gender_cols = ['gender']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:57:44.207565572Z",
     "start_time": "2023-12-31T03:57:44.202177544Z"
    },
    "collapsed": false
   },
   "source": [
    "\n",
    "Choose a single column from the candidate columns that record age and gender information respectively.\n",
    "If no column meets the requirement, keep 'age_col' or 'gender_col' to None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:22:26.108535524Z",
     "start_time": "2024-02-27T00:22:26.101435803Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "age_col = 'age_at_initial_pathologic_diagnosis'\n",
    "gender_col = 'gender'\n",
    "selected_clinical_data = xena_select_clinical_features(clinical_data, TRAIT, age_col=age_col, gender_col=gender_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.602458026Z",
     "start_time": "2024-02-27T00:22:27.839636769Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-N7-A4Y5-01</th>\n",
       "      <th>TCGA-N5-A4RM-01</th>\n",
       "      <th>TCGA-N9-A4Q8-01</th>\n",
       "      <th>TCGA-N9-A4Q7-01</th>\n",
       "      <th>TCGA-N6-A4VD-01</th>\n",
       "      <th>TCGA-N5-A4RT-01</th>\n",
       "      <th>TCGA-N9-A4Q1-01</th>\n",
       "      <th>TCGA-N5-A4RD-01</th>\n",
       "      <th>TCGA-N8-A4PL-01</th>\n",
       "      <th>TCGA-ND-A4WC-01</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-ND-A4WA-01</th>\n",
       "      <th>TCGA-N8-A56S-01</th>\n",
       "      <th>TCGA-ND-A4WF-01</th>\n",
       "      <th>TCGA-N5-A4RJ-01</th>\n",
       "      <th>TCGA-NA-A4QX-01</th>\n",
       "      <th>TCGA-N5-A4RS-01</th>\n",
       "      <th>TCGA-N5-A4RO-01</th>\n",
       "      <th>TCGA-NF-A5CP-01</th>\n",
       "      <th>TCGA-N6-A4VC-01</th>\n",
       "      <th>TCGA-N5-A4RU-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <td>0.159808</td>\n",
       "      <td>-0.342892</td>\n",
       "      <td>-2.561692</td>\n",
       "      <td>0.814708</td>\n",
       "      <td>-0.747992</td>\n",
       "      <td>-1.216492</td>\n",
       "      <td>-1.183192</td>\n",
       "      <td>-1.075592</td>\n",
       "      <td>-0.139592</td>\n",
       "      <td>-0.308092</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.399792</td>\n",
       "      <td>0.710908</td>\n",
       "      <td>-0.276192</td>\n",
       "      <td>0.275708</td>\n",
       "      <td>-0.697392</td>\n",
       "      <td>-0.258992</td>\n",
       "      <td>-1.682292</td>\n",
       "      <td>-0.213692</td>\n",
       "      <td>-1.378792</td>\n",
       "      <td>-1.113392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIF3A</th>\n",
       "      <td>8.116374</td>\n",
       "      <td>3.251374</td>\n",
       "      <td>3.964374</td>\n",
       "      <td>2.101974</td>\n",
       "      <td>1.745474</td>\n",
       "      <td>4.559774</td>\n",
       "      <td>6.436574</td>\n",
       "      <td>5.400374</td>\n",
       "      <td>5.781974</td>\n",
       "      <td>2.870174</td>\n",
       "      <td>...</td>\n",
       "      <td>5.370274</td>\n",
       "      <td>7.416774</td>\n",
       "      <td>5.491974</td>\n",
       "      <td>5.896174</td>\n",
       "      <td>5.058874</td>\n",
       "      <td>4.639074</td>\n",
       "      <td>5.189674</td>\n",
       "      <td>0.857074</td>\n",
       "      <td>1.616374</td>\n",
       "      <td>4.505174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF17</th>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.023165</td>\n",
       "      <td>0.337765</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.006465</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.372565</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.531035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF10</th>\n",
       "      <td>0.491128</td>\n",
       "      <td>0.449828</td>\n",
       "      <td>-0.213572</td>\n",
       "      <td>0.247028</td>\n",
       "      <td>0.142628</td>\n",
       "      <td>-0.295572</td>\n",
       "      <td>-0.819172</td>\n",
       "      <td>-0.327472</td>\n",
       "      <td>0.092528</td>\n",
       "      <td>0.082328</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.056772</td>\n",
       "      <td>0.087828</td>\n",
       "      <td>0.532928</td>\n",
       "      <td>-0.087372</td>\n",
       "      <td>0.173328</td>\n",
       "      <td>-1.143872</td>\n",
       "      <td>-0.096972</td>\n",
       "      <td>-0.306672</td>\n",
       "      <td>0.606228</td>\n",
       "      <td>0.110928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF11</th>\n",
       "      <td>0.221222</td>\n",
       "      <td>-0.508578</td>\n",
       "      <td>-0.239378</td>\n",
       "      <td>-1.285978</td>\n",
       "      <td>-0.141278</td>\n",
       "      <td>0.112322</td>\n",
       "      <td>0.777922</td>\n",
       "      <td>-0.623978</td>\n",
       "      <td>-1.300578</td>\n",
       "      <td>-0.339078</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.258378</td>\n",
       "      <td>-0.149078</td>\n",
       "      <td>-0.143578</td>\n",
       "      <td>-0.719378</td>\n",
       "      <td>0.129022</td>\n",
       "      <td>-0.635478</td>\n",
       "      <td>-1.139278</td>\n",
       "      <td>-0.246478</td>\n",
       "      <td>0.172722</td>\n",
       "      <td>-0.160778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNGT1</th>\n",
       "      <td>2.162310</td>\n",
       "      <td>2.528110</td>\n",
       "      <td>-1.281390</td>\n",
       "      <td>2.062210</td>\n",
       "      <td>-0.500290</td>\n",
       "      <td>1.902010</td>\n",
       "      <td>2.914810</td>\n",
       "      <td>3.925510</td>\n",
       "      <td>1.604410</td>\n",
       "      <td>1.498710</td>\n",
       "      <td>...</td>\n",
       "      <td>3.084910</td>\n",
       "      <td>2.572910</td>\n",
       "      <td>3.055810</td>\n",
       "      <td>1.251910</td>\n",
       "      <td>3.911510</td>\n",
       "      <td>4.970610</td>\n",
       "      <td>3.990210</td>\n",
       "      <td>1.539910</td>\n",
       "      <td>2.313210</td>\n",
       "      <td>0.151610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TULP3</th>\n",
       "      <td>1.156223</td>\n",
       "      <td>0.980423</td>\n",
       "      <td>0.125923</td>\n",
       "      <td>-0.115377</td>\n",
       "      <td>0.394023</td>\n",
       "      <td>0.373923</td>\n",
       "      <td>1.008823</td>\n",
       "      <td>0.600623</td>\n",
       "      <td>1.977023</td>\n",
       "      <td>0.193523</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.100777</td>\n",
       "      <td>0.737523</td>\n",
       "      <td>0.622323</td>\n",
       "      <td>0.508223</td>\n",
       "      <td>1.000123</td>\n",
       "      <td>1.038223</td>\n",
       "      <td>0.882223</td>\n",
       "      <td>0.320623</td>\n",
       "      <td>1.178323</td>\n",
       "      <td>-0.238477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL6B</th>\n",
       "      <td>-1.218927</td>\n",
       "      <td>-0.407627</td>\n",
       "      <td>-1.056927</td>\n",
       "      <td>-1.821127</td>\n",
       "      <td>-1.103927</td>\n",
       "      <td>-0.457927</td>\n",
       "      <td>-0.592327</td>\n",
       "      <td>-0.892127</td>\n",
       "      <td>-3.291227</td>\n",
       "      <td>-0.344127</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.937927</td>\n",
       "      <td>0.270373</td>\n",
       "      <td>0.005173</td>\n",
       "      <td>0.983573</td>\n",
       "      <td>-0.433727</td>\n",
       "      <td>-1.194427</td>\n",
       "      <td>-2.458227</td>\n",
       "      <td>0.992073</td>\n",
       "      <td>-1.211927</td>\n",
       "      <td>0.821373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSTK1</th>\n",
       "      <td>-1.559495</td>\n",
       "      <td>-0.638195</td>\n",
       "      <td>-0.807795</td>\n",
       "      <td>1.332405</td>\n",
       "      <td>-0.030595</td>\n",
       "      <td>-1.583695</td>\n",
       "      <td>-1.032395</td>\n",
       "      <td>-0.581895</td>\n",
       "      <td>-0.764895</td>\n",
       "      <td>0.187505</td>\n",
       "      <td>...</td>\n",
       "      <td>0.145405</td>\n",
       "      <td>-0.545095</td>\n",
       "      <td>-0.330595</td>\n",
       "      <td>0.058705</td>\n",
       "      <td>-1.419095</td>\n",
       "      <td>-0.349695</td>\n",
       "      <td>-1.139895</td>\n",
       "      <td>-0.443995</td>\n",
       "      <td>-0.818895</td>\n",
       "      <td>-1.583695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELP</th>\n",
       "      <td>-2.305133</td>\n",
       "      <td>-0.953433</td>\n",
       "      <td>-3.142633</td>\n",
       "      <td>-1.671733</td>\n",
       "      <td>-2.314333</td>\n",
       "      <td>-2.458333</td>\n",
       "      <td>-0.407733</td>\n",
       "      <td>-3.565233</td>\n",
       "      <td>-4.636333</td>\n",
       "      <td>-0.402033</td>\n",
       "      <td>...</td>\n",
       "      <td>0.632667</td>\n",
       "      <td>-0.679133</td>\n",
       "      <td>-4.959633</td>\n",
       "      <td>0.255367</td>\n",
       "      <td>-1.856033</td>\n",
       "      <td>-1.530333</td>\n",
       "      <td>-4.626233</td>\n",
       "      <td>0.696067</td>\n",
       "      <td>-3.610933</td>\n",
       "      <td>-3.881633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17339 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TCGA-N7-A4Y5-01  TCGA-N5-A4RM-01  TCGA-N9-A4Q8-01  TCGA-N9-A4Q7-01  \\\n",
       "ARHGEF10L         0.159808        -0.342892        -2.561692         0.814708   \n",
       "HIF3A             8.116374         3.251374         3.964374         2.101974   \n",
       "RNF17            -0.531035        -0.531035        -0.531035        -0.531035   \n",
       "RNF10             0.491128         0.449828        -0.213572         0.247028   \n",
       "RNF11             0.221222        -0.508578        -0.239378        -1.285978   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             2.162310         2.528110        -1.281390         2.062210   \n",
       "TULP3             1.156223         0.980423         0.125923        -0.115377   \n",
       "BCL6B            -1.218927        -0.407627        -1.056927        -1.821127   \n",
       "GSTK1            -1.559495        -0.638195        -0.807795         1.332405   \n",
       "SELP             -2.305133        -0.953433        -3.142633        -1.671733   \n",
       "\n",
       "           TCGA-N6-A4VD-01  TCGA-N5-A4RT-01  TCGA-N9-A4Q1-01  TCGA-N5-A4RD-01  \\\n",
       "ARHGEF10L        -0.747992        -1.216492        -1.183192        -1.075592   \n",
       "HIF3A             1.745474         4.559774         6.436574         5.400374   \n",
       "RNF17            -0.531035        -0.531035         0.023165         0.337765   \n",
       "RNF10             0.142628        -0.295572        -0.819172        -0.327472   \n",
       "RNF11            -0.141278         0.112322         0.777922        -0.623978   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1            -0.500290         1.902010         2.914810         3.925510   \n",
       "TULP3             0.394023         0.373923         1.008823         0.600623   \n",
       "BCL6B            -1.103927        -0.457927        -0.592327        -0.892127   \n",
       "GSTK1            -0.030595        -1.583695        -1.032395        -0.581895   \n",
       "SELP             -2.314333        -2.458333        -0.407733        -3.565233   \n",
       "\n",
       "           TCGA-N8-A4PL-01  TCGA-ND-A4WC-01  ...  TCGA-ND-A4WA-01  \\\n",
       "ARHGEF10L        -0.139592        -0.308092  ...        -0.399792   \n",
       "HIF3A             5.781974         2.870174  ...         5.370274   \n",
       "RNF17            -0.531035         0.006465  ...        -0.531035   \n",
       "RNF10             0.092528         0.082328  ...        -0.056772   \n",
       "RNF11            -1.300578        -0.339078  ...        -0.258378   \n",
       "...                    ...              ...  ...              ...   \n",
       "GNGT1             1.604410         1.498710  ...         3.084910   \n",
       "TULP3             1.977023         0.193523  ...        -0.100777   \n",
       "BCL6B            -3.291227        -0.344127  ...        -0.937927   \n",
       "GSTK1            -0.764895         0.187505  ...         0.145405   \n",
       "SELP             -4.636333        -0.402033  ...         0.632667   \n",
       "\n",
       "           TCGA-N8-A56S-01  TCGA-ND-A4WF-01  TCGA-N5-A4RJ-01  TCGA-NA-A4QX-01  \\\n",
       "ARHGEF10L         0.710908        -0.276192         0.275708        -0.697392   \n",
       "HIF3A             7.416774         5.491974         5.896174         5.058874   \n",
       "RNF17            -0.531035        -0.531035         0.372565        -0.531035   \n",
       "RNF10             0.087828         0.532928        -0.087372         0.173328   \n",
       "RNF11            -0.149078        -0.143578        -0.719378         0.129022   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             2.572910         3.055810         1.251910         3.911510   \n",
       "TULP3             0.737523         0.622323         0.508223         1.000123   \n",
       "BCL6B             0.270373         0.005173         0.983573        -0.433727   \n",
       "GSTK1            -0.545095        -0.330595         0.058705        -1.419095   \n",
       "SELP             -0.679133        -4.959633         0.255367        -1.856033   \n",
       "\n",
       "           TCGA-N5-A4RS-01  TCGA-N5-A4RO-01  TCGA-NF-A5CP-01  TCGA-N6-A4VC-01  \\\n",
       "ARHGEF10L        -0.258992        -1.682292        -0.213692        -1.378792   \n",
       "HIF3A             4.639074         5.189674         0.857074         1.616374   \n",
       "RNF17            -0.531035        -0.531035        -0.531035        -0.531035   \n",
       "RNF10            -1.143872        -0.096972        -0.306672         0.606228   \n",
       "RNF11            -0.635478        -1.139278        -0.246478         0.172722   \n",
       "...                    ...              ...              ...              ...   \n",
       "GNGT1             4.970610         3.990210         1.539910         2.313210   \n",
       "TULP3             1.038223         0.882223         0.320623         1.178323   \n",
       "BCL6B            -1.194427        -2.458227         0.992073        -1.211927   \n",
       "GSTK1            -0.349695        -1.139895        -0.443995        -0.818895   \n",
       "SELP             -1.530333        -4.626233         0.696067        -3.610933   \n",
       "\n",
       "           TCGA-N5-A4RU-01  \n",
       "ARHGEF10L        -1.113392  \n",
       "HIF3A             4.505174  \n",
       "RNF17            -0.531035  \n",
       "RNF10             0.110928  \n",
       "RNF11            -0.160778  \n",
       "...                    ...  \n",
       "GNGT1             0.151610  \n",
       "TULP3            -0.238477  \n",
       "BCL6B             0.821373  \n",
       "GSTK1            -1.583695  \n",
       "SELP             -3.881633  \n",
       "\n",
       "[17339 rows x 57 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.720144304Z",
     "start_time": "2024-02-27T00:23:16.643071320Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uterine Carcinosarcoma</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>ARHGEF10L</th>\n",
       "      <th>HIF3A</th>\n",
       "      <th>RNF17</th>\n",
       "      <th>RNF10</th>\n",
       "      <th>RNF11</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>GTF2IP1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC7A10</th>\n",
       "      <th>PLA2G2C</th>\n",
       "      <th>TULP2</th>\n",
       "      <th>NPY5R</th>\n",
       "      <th>GNGT2</th>\n",
       "      <th>GNGT1</th>\n",
       "      <th>TULP3</th>\n",
       "      <th>BCL6B</th>\n",
       "      <th>GSTK1</th>\n",
       "      <th>SELP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-N5-A4R8-01</th>\n",
       "      <td>1</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.957792</td>\n",
       "      <td>6.814974</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>-0.200472</td>\n",
       "      <td>-0.535378</td>\n",
       "      <td>-1.42261</td>\n",
       "      <td>-0.539194</td>\n",
       "      <td>...</td>\n",
       "      <td>5.055914</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.017922</td>\n",
       "      <td>3.273883</td>\n",
       "      <td>-0.262233</td>\n",
       "      <td>1.95401</td>\n",
       "      <td>1.290723</td>\n",
       "      <td>-0.852727</td>\n",
       "      <td>-2.000395</td>\n",
       "      <td>-5.091733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-N5-A4RA-01</th>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.436292</td>\n",
       "      <td>6.814374</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.387528</td>\n",
       "      <td>-0.035178</td>\n",
       "      <td>-1.14681</td>\n",
       "      <td>-0.246094</td>\n",
       "      <td>...</td>\n",
       "      <td>6.600414</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>0.234322</td>\n",
       "      <td>-1.013317</td>\n",
       "      <td>-1.696233</td>\n",
       "      <td>0.02041</td>\n",
       "      <td>0.931723</td>\n",
       "      <td>-2.614627</td>\n",
       "      <td>-0.800295</td>\n",
       "      <td>-5.525533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-N5-A4RD-01</th>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.075592</td>\n",
       "      <td>5.400374</td>\n",
       "      <td>0.337765</td>\n",
       "      <td>-0.327472</td>\n",
       "      <td>-0.623978</td>\n",
       "      <td>-0.76911</td>\n",
       "      <td>0.301706</td>\n",
       "      <td>...</td>\n",
       "      <td>6.132114</td>\n",
       "      <td>0.412218</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>1.260083</td>\n",
       "      <td>0.190467</td>\n",
       "      <td>3.92551</td>\n",
       "      <td>0.600623</td>\n",
       "      <td>-0.892127</td>\n",
       "      <td>-0.581895</td>\n",
       "      <td>-3.565233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-N5-A4RF-01</th>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.616892</td>\n",
       "      <td>4.298374</td>\n",
       "      <td>-0.531035</td>\n",
       "      <td>0.068628</td>\n",
       "      <td>-0.272078</td>\n",
       "      <td>-1.90441</td>\n",
       "      <td>-0.123494</td>\n",
       "      <td>...</td>\n",
       "      <td>5.434814</td>\n",
       "      <td>1.727818</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>-1.587117</td>\n",
       "      <td>2.072467</td>\n",
       "      <td>3.56111</td>\n",
       "      <td>0.549123</td>\n",
       "      <td>0.040173</td>\n",
       "      <td>-0.607895</td>\n",
       "      <td>-3.091133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-N5-A4RJ-01</th>\n",
       "      <td>1</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0.275708</td>\n",
       "      <td>5.896174</td>\n",
       "      <td>0.372565</td>\n",
       "      <td>-0.087372</td>\n",
       "      <td>-0.719378</td>\n",
       "      <td>-1.19391</td>\n",
       "      <td>-0.032194</td>\n",
       "      <td>...</td>\n",
       "      <td>5.051014</td>\n",
       "      <td>-0.086682</td>\n",
       "      <td>-0.748878</td>\n",
       "      <td>0.710983</td>\n",
       "      <td>0.398067</td>\n",
       "      <td>1.25191</td>\n",
       "      <td>0.508223</td>\n",
       "      <td>0.983573</td>\n",
       "      <td>0.058705</td>\n",
       "      <td>0.255367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 17342 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Uterine Carcinosarcoma  Age  Gender  ARHGEF10L     HIF3A  \\\n",
       "sampleID                                                                    \n",
       "TCGA-N5-A4R8-01                       1   65       0  -0.957792  6.814974   \n",
       "TCGA-N5-A4RA-01                       1   63       0  -0.436292  6.814374   \n",
       "TCGA-N5-A4RD-01                       1   69       0  -1.075592  5.400374   \n",
       "TCGA-N5-A4RF-01                       1   68       0  -0.616892  4.298374   \n",
       "TCGA-N5-A4RJ-01                       1   61       0   0.275708  5.896174   \n",
       "\n",
       "                    RNF17     RNF10     RNF11    RNF13   GTF2IP1  ...  \\\n",
       "sampleID                                                          ...   \n",
       "TCGA-N5-A4R8-01 -0.531035 -0.200472 -0.535378 -1.42261 -0.539194  ...   \n",
       "TCGA-N5-A4RA-01 -0.531035  0.387528 -0.035178 -1.14681 -0.246094  ...   \n",
       "TCGA-N5-A4RD-01  0.337765 -0.327472 -0.623978 -0.76911  0.301706  ...   \n",
       "TCGA-N5-A4RF-01 -0.531035  0.068628 -0.272078 -1.90441 -0.123494  ...   \n",
       "TCGA-N5-A4RJ-01  0.372565 -0.087372 -0.719378 -1.19391 -0.032194  ...   \n",
       "\n",
       "                  SLC7A10   PLA2G2C     TULP2     NPY5R     GNGT2    GNGT1  \\\n",
       "sampleID                                                                     \n",
       "TCGA-N5-A4R8-01  5.055914 -0.086682  0.017922  3.273883 -0.262233  1.95401   \n",
       "TCGA-N5-A4RA-01  6.600414 -0.086682  0.234322 -1.013317 -1.696233  0.02041   \n",
       "TCGA-N5-A4RD-01  6.132114  0.412218 -0.748878  1.260083  0.190467  3.92551   \n",
       "TCGA-N5-A4RF-01  5.434814  1.727818 -0.748878 -1.587117  2.072467  3.56111   \n",
       "TCGA-N5-A4RJ-01  5.051014 -0.086682 -0.748878  0.710983  0.398067  1.25191   \n",
       "\n",
       "                    TULP3     BCL6B     GSTK1      SELP  \n",
       "sampleID                                                 \n",
       "TCGA-N5-A4R8-01  1.290723 -0.852727 -2.000395 -5.091733  \n",
       "TCGA-N5-A4RA-01  0.931723 -2.614627 -0.800295 -5.525533  \n",
       "TCGA-N5-A4RD-01  0.600623 -0.892127 -0.581895 -3.565233  \n",
       "TCGA-N5-A4RF-01  0.549123  0.040173 -0.607895 -3.091133  \n",
       "TCGA-N5-A4RJ-01  0.508223  0.983573  0.058705  0.255367  \n",
       "\n",
       "[5 rows x 17342 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = selected_clinical_data.join(genetic_data.T).dropna()\n",
    "merged_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:16.731303752Z",
     "start_time": "2024-02-27T00:23:16.717316474Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 57 samples.\n",
      "For the feature 'Uterine Carcinosarcoma', the least common label is '1' with 57 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Uterine Carcinosarcoma' in this dataset is severely biased.\n",
      "\n",
      "Quartiles for 'Age':\n",
      "  25%: 62.0\n",
      "  50% (Median): 68.0\n",
      "  75%: 76.0\n",
      "Min: 51\n",
      "Max: 90\n",
      "The distribution of the feature 'Age' in this dataset is fine.\n",
      "\n",
      "For the feature 'Gender', the least common label is '0' with 57 occurrences. This represents 100.00% of the dataset.\n",
      "The distribution of the feature 'Gender' in this dataset is severely biased.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merge_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.252387093Z",
     "start_time": "2024-02-27T00:23:16.722533204Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merge_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302271525Z",
     "start_time": "2024-02-27T00:23:30.295066253Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A new JSON file was created at: /Users/legion/Desktop/Courses/IS389/output2\\Jiayi\\Uterine-Carcinosarcoma\\cohort_info.json\n"
     ]
    }
   ],
   "source": [
    "save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-28T21:28:38.028916303Z",
     "start_time": "2023-12-28T21:28:38.016245426Z"
    },
    "collapsed": false
   },
   "source": [
    "## 2.2. The GEO dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302587016Z",
     "start_time": "2024-02-27T00:23:30.295278811Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSE16680', 'GSE32507', 'GSE36133', 'GSE36138', 'GSE36139', 'GSE68950']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = 'GEO'\n",
    "trait_subdir = \"Uterine-Carcinosarcoma\"\n",
    "\n",
    "trait_path = os.path.join(DATA_ROOT, dataset, trait_subdir)\n",
    "os.listdir(trait_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:39:42.634870142Z",
     "start_time": "2023-12-31T03:39:42.534093295Z"
    },
    "collapsed": false
   },
   "source": [
    "Repeat the below steps for all the accession numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.302864515Z",
     "start_time": "2024-02-27T00:23:30.295388276Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE16680\\\\GSE16680_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE16680\\\\GSE16680_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Biased\n",
    "cohort = accession_num = \"GSE16680\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE32507\\\\GSE32507_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE32507\\\\GSE32507_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Finished\n",
    "cohort = accession_num = \"GSE32507\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE36133\\\\GSE36133_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE36133\\\\GSE36133_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE36133\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE36138\\\\GSE36138_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE36138\\\\GSE36138_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MemoryError\n",
    "cohort = accession_num = \"GSE36138\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE36139\\\\GSE36139_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE36139\\\\GSE36139-GPL15308_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No gene mapping\n",
    "cohort = accession_num = \"GSE36139\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE68950\\\\GSE68950_family.soft.gz',\n",
       " '/Users/legion/Desktop/Courses/IS389/data\\\\GEO\\\\Uterine-Carcinosarcoma\\\\GSE68950\\\\GSE68950_series_matrix.txt.gz')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# No obvious trait\n",
    "cohort = accession_num = \"GSE68950\"\n",
    "cohort_dir = os.path.join(trait_path, accession_num)\n",
    "soft_file, matrix_file = geo_get_relevant_filepaths(cohort_dir)\n",
    "soft_file, matrix_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### Initial filtering and clinical data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303301102Z",
     "start_time": "2024-02-27T00:23:30.295491459Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Series_title\t\"caArray_golub-00327: Sanger cell line Affymetrix gene expression project\"\n",
      "!Series_summary\t\"The microarray gene expression pattern was studied using 798 different cancer cell lines. The cancer cell lines are obtained from different centers. Annotation information were provided in the supplementary file.\"\n",
      "!Series_overall_design\t\"golub-00327\"\n",
      "!Series_overall_design\t\"Assay Type: Gene Expression\"\n",
      "!Series_overall_design\t\"Provider: Affymetrix\"\n",
      "!Series_overall_design\t\"Array Designs: HT_HG-U133A\"\n",
      "!Series_overall_design\t\"Organism: Homo sapiens (ncbitax)\"\n",
      "!Series_overall_design\t\"Tissue Sites: leukemia, Urinary tract, Lung, BiliaryTract, Autonomic Ganglion, Thyroid gland, Stomach, Breast, Pancreas, Head and Neck, Lymphoma, Colorectal, Placenta, Liver, Brain, Bone, pleura, Skin, endometrium, Ovary, cervix, Oesophagus, Connective and Soft Tissue, Muscle, Kidney, Prostate, Adrenal Gland, Eye, Testis, Smooth Muscle Tissue, Vulva, Unknow\"\n",
      "!Series_overall_design\t\"Material Types: cell, synthetic_RNA, whole_organism, total_RNA, BVG\"\n",
      "!Series_overall_design\t\"Disease States: M3 acute myeloid leukemia, hairy cell leukemia, transitional cell carcinoma, Adenocarcinoma, B cell lymphoma unspecified, Acute Lymphoblastic Leukemia, blast phase chronic myeloid leukemia, Carcinoma, M6 acute myeloid leukemia, Neuroblastoma, follicular carcinoma, ductal carcinoma, Burkitt Lymphoma, Squamous Cell Carcinoma, M5 acute myeloid leukemia, Mycosis Fungoides and Sezary Syndrome, Acute T-Cell Lymphoblastic Leukemia, Adult T-Cell Leukemia/Lymphoma, M2 Therapy-Related Myeloid Neoplasm, Choriocarcinoma, Plasma Cell Myeloma, Hepatocellular Carcinoma, anaplastic large cell lymphoma, primitive neuroectodermal tumor-medulloblastoma, M4 acute myeloid leukemia, B Acute Lymphoblastic Leukemia, Acute Leukemia of Ambiguous Lineage, Osteosarcoma, Hodgkin Lymphoma, Mesothelioma, chondrosarcoma, Glioblastoma Multiforme, Malignant Melanoma, carcinosarcoma-malignant mesodermal mixed tumor, bronchioloalveolar adenocarcinoma, chronic lymphocytic leukemia-small lymphocytic lymphoma, micropapillary carcinoma, diffuse large B cell lymphoma, myelodysplastic syndrome, giant cell carcinoma, teratoma, multipotential sarcoma, Small Cell Carcinoma, ASTROCYTOMA, Fibrosarcoma, mucoepidermoid carcinoma, Rhabdomyosarcoma, L1 Acute T-Cell Lymphoblastic Leukemia, Glioma, Anaplastic Astrocytoma, Non-small cell carcinoma, Large Cell Carcinoma, mucinous carcinoma, Acute Myeloid Leukemia, malignant fibrous histiocytoma-pleomorphic sarcoma, clear cell carcinoma, B cell lymphoma unspecified, Anaplastic Carcinoma, Ewings sarcoma-peripheral primitive neuroectodermal tumor, undifferentiated carcinoma, Sarcoma, Embryonal Rhabdomyosarcoma, epithelioid sarcoma, renal cell carcinoma, carcinoid-endocrine tumor, Synovial Sarcoma, lymphoid neoplasm, rhabdoid tumor, Refractory Anemia with Excess Blasts, Liposarcoma, biphasic mesothelioma, adrenal cortical carcinoma, adenosquamous carcinoma, L2 Acute T-Cell Lymphoblastic Leukemia, chronic myeloid leukemia, Micropapillary Serous Carcinoma, desmoplastic, acute leukemia, Retinoblastoma, teratocarcinoma, clear cell renal cell carcinoma, Follicular Lymphoma, Wilms Tumor, M7 acute myeloid leukemia, gliosarcoma, embryonal carcinoma, Leiomyosarcoma, medullary carcinoma, granulosa cell tumor, papillary carcinoma, NS Acute Lymphoblastic Leukemia, papillary transitional cell carcinoma, small cell adenocarcinoma, epithelial dysplasia, hyperplasia, tubular adenocarcinoma, metaplasia, papillary ductal carcinoma, chronic eosinophilic leukemia-hypereosinophilic syndrome, #N/A, malignant trichilemmal cyst, Medullary Breast Carcinoma, L2 Acute Lymphoblastic Leukemia, Osteoblastic Osteosarcoma\"\n"
     ]
    }
   ],
   "source": [
    "from utils import *\n",
    "background_prefixes = ['!Series_title', '!Series_summary', '!Series_overall_design']\n",
    "clinical_prefixes = ['!Sample_geo_accession', '!Sample_characteristics_ch1']    \n",
    "\n",
    "background_info, clinical_data = get_background_and_clinical_data(matrix_file, background_prefixes, clinical_prefixes)\n",
    "print(background_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.303593599Z",
     "start_time": "2024-02-27T00:23:30.299747893Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th>GSM1687570</th>\n",
       "      <th>GSM1687571</th>\n",
       "      <th>GSM1687572</th>\n",
       "      <th>GSM1687573</th>\n",
       "      <th>GSM1687574</th>\n",
       "      <th>GSM1687575</th>\n",
       "      <th>GSM1687576</th>\n",
       "      <th>GSM1687577</th>\n",
       "      <th>GSM1687578</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1688358</th>\n",
       "      <th>GSM1688359</th>\n",
       "      <th>GSM1688360</th>\n",
       "      <th>GSM1688361</th>\n",
       "      <th>GSM1688362</th>\n",
       "      <th>GSM1688363</th>\n",
       "      <th>GSM1688364</th>\n",
       "      <th>GSM1688365</th>\n",
       "      <th>GSM1688366</th>\n",
       "      <th>GSM1688367</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cosmic id: 924101</td>\n",
       "      <td>cosmic id: 906800</td>\n",
       "      <td>cosmic id: 687452</td>\n",
       "      <td>cosmic id: 924100</td>\n",
       "      <td>cosmic id: 910924</td>\n",
       "      <td>cosmic id: 906798</td>\n",
       "      <td>cosmic id: 906797</td>\n",
       "      <td>cosmic id: 906797</td>\n",
       "      <td>cosmic id: 910922</td>\n",
       "      <td>...</td>\n",
       "      <td>cosmic id: 909781</td>\n",
       "      <td>cosmic id: 909782</td>\n",
       "      <td>cosmic id: 909782</td>\n",
       "      <td>cosmic id: 909784</td>\n",
       "      <td>cosmic id: 909785</td>\n",
       "      <td>cosmic id: 909785</td>\n",
       "      <td>cosmic id: 909904</td>\n",
       "      <td>cosmic id: 909905</td>\n",
       "      <td>cosmic id: 687592</td>\n",
       "      <td>cosmic id: 909907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease state: L2 Acute Lymphoblastic Leukemia</td>\n",
       "      <td>disease state: NS Acute Lymphoblastic Leukemia</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: adenocarcinoma</td>\n",
       "      <td>disease state: adenocarcinoma</td>\n",
       "      <td>disease state: transitional cell carcinoma</td>\n",
       "      <td>disease state: transitional cell carcinoma</td>\n",
       "      <td>disease state: transitional cell carcinoma</td>\n",
       "      <td>disease state: clear cell renal cell carcinoma</td>\n",
       "      <td>...</td>\n",
       "      <td>disease state: renal cell carcinoma</td>\n",
       "      <td>disease state: retinoblastoma</td>\n",
       "      <td>disease state: retinoblastoma</td>\n",
       "      <td>disease state: malignant melanoma</td>\n",
       "      <td>disease state: follicular lymphoma</td>\n",
       "      <td>disease state: follicular lymphoma</td>\n",
       "      <td>disease state: carcinoma</td>\n",
       "      <td>disease state: glioblastoma multiforme</td>\n",
       "      <td>disease state: glioblastoma multiforme</td>\n",
       "      <td>disease state: ductal carcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>disease location: Hematopoietic and Lymphoid T...</td>\n",
       "      <td>disease location: Hematopoietic and Lymphoid T...</td>\n",
       "      <td>disease location: bladder</td>\n",
       "      <td>disease location: prostate</td>\n",
       "      <td>disease location: stomach</td>\n",
       "      <td>disease location: ureter</td>\n",
       "      <td>disease location: bladder</td>\n",
       "      <td>disease location: bladder</td>\n",
       "      <td>disease location: kidney</td>\n",
       "      <td>...</td>\n",
       "      <td>disease location: kidney</td>\n",
       "      <td>disease location: retina</td>\n",
       "      <td>disease location: retina</td>\n",
       "      <td>disease location: skin</td>\n",
       "      <td>disease location: lymph node</td>\n",
       "      <td>disease location: lymph node</td>\n",
       "      <td>disease location: pancreas</td>\n",
       "      <td>disease location: brain</td>\n",
       "      <td>disease location: temporal lobe</td>\n",
       "      <td>disease location: breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>organism part: Leukemia</td>\n",
       "      <td>organism part: Leukemia</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Prostate</td>\n",
       "      <td>organism part: Stomach</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Urinary tract</td>\n",
       "      <td>organism part: Kidney</td>\n",
       "      <td>...</td>\n",
       "      <td>organism part: Kidney</td>\n",
       "      <td>organism part: Eye</td>\n",
       "      <td>organism part: Eye</td>\n",
       "      <td>organism part: Skin</td>\n",
       "      <td>organism part: Lymphoma</td>\n",
       "      <td>organism part: Lymphoma</td>\n",
       "      <td>organism part: Pancreas</td>\n",
       "      <td>organism part: Brain</td>\n",
       "      <td>organism part: Brain</td>\n",
       "      <td>organism part: Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>sample: 736</td>\n",
       "      <td>sample: 494</td>\n",
       "      <td>sample: 7</td>\n",
       "      <td>sample: 746</td>\n",
       "      <td>sample: 439</td>\n",
       "      <td>sample: 168</td>\n",
       "      <td>sample: 152</td>\n",
       "      <td>sample: 37</td>\n",
       "      <td>sample: 450</td>\n",
       "      <td>...</td>\n",
       "      <td>sample: 470</td>\n",
       "      <td>sample: 246</td>\n",
       "      <td>sample: 246</td>\n",
       "      <td>sample: 714</td>\n",
       "      <td>sample: 482</td>\n",
       "      <td>sample: 49</td>\n",
       "      <td>sample: 234</td>\n",
       "      <td>sample: 41</td>\n",
       "      <td>sample: 397</td>\n",
       "      <td>sample: 726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>cell line code: 749</td>\n",
       "      <td>cell line code: 493</td>\n",
       "      <td>cell line code: 505</td>\n",
       "      <td>cell line code: 760</td>\n",
       "      <td>cell line code: 437</td>\n",
       "      <td>cell line code: 151</td>\n",
       "      <td>cell line code: 134</td>\n",
       "      <td>cell line code: 134</td>\n",
       "      <td>cell line code: 449</td>\n",
       "      <td>...</td>\n",
       "      <td>cell line code: 469</td>\n",
       "      <td>cell line code: 231</td>\n",
       "      <td>cell line code: 231</td>\n",
       "      <td>cell line code: 727</td>\n",
       "      <td>cell line code: 481</td>\n",
       "      <td>cell line code: 481</td>\n",
       "      <td>cell line code: 219</td>\n",
       "      <td>cell line code: 401</td>\n",
       "      <td>cell line code: 390</td>\n",
       "      <td>cell line code: 738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: Unspecified</td>\n",
       "      <td>...</td>\n",
       "      <td>supplier: HSRRB</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: DSMZ</td>\n",
       "      <td>supplier: HSRRB</td>\n",
       "      <td>supplier: HSRRB</td>\n",
       "      <td>supplier: ATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>...</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 2</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "      <td>affy_batch: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>!Sample_characteristics_ch1</td>\n",
       "      <td>crna plate: 8</td>\n",
       "      <td>crna plate: 6</td>\n",
       "      <td>crna plate: 11</td>\n",
       "      <td>crna plate: 8</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>crna plate: 2</td>\n",
       "      <td>crna plate: 2</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>...</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>crna plate: 3</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 8</td>\n",
       "      <td>crna plate: 5</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 3</td>\n",
       "      <td>crna plate: 12</td>\n",
       "      <td>crna plate: 4</td>\n",
       "      <td>crna plate: 8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 799 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         !Sample_geo_accession  \\\n",
       "0  !Sample_characteristics_ch1   \n",
       "1  !Sample_characteristics_ch1   \n",
       "2  !Sample_characteristics_ch1   \n",
       "3  !Sample_characteristics_ch1   \n",
       "4  !Sample_characteristics_ch1   \n",
       "5  !Sample_characteristics_ch1   \n",
       "6  !Sample_characteristics_ch1   \n",
       "7  !Sample_characteristics_ch1   \n",
       "8  !Sample_characteristics_ch1   \n",
       "\n",
       "                                          GSM1687570  \\\n",
       "0                                  cosmic id: 924101   \n",
       "1     disease state: L2 Acute Lymphoblastic Leukemia   \n",
       "2  disease location: Hematopoietic and Lymphoid T...   \n",
       "3                            organism part: Leukemia   \n",
       "4                                        sample: 736   \n",
       "5                                cell line code: 749   \n",
       "6                                     supplier: DSMZ   \n",
       "7                                      affy_batch: 1   \n",
       "8                                      crna plate: 8   \n",
       "\n",
       "                                          GSM1687571  \\\n",
       "0                                  cosmic id: 906800   \n",
       "1     disease state: NS Acute Lymphoblastic Leukemia   \n",
       "2  disease location: Hematopoietic and Lymphoid T...   \n",
       "3                            organism part: Leukemia   \n",
       "4                                        sample: 494   \n",
       "5                                cell line code: 493   \n",
       "6                                     supplier: DSMZ   \n",
       "7                                      affy_batch: 1   \n",
       "8                                      crna plate: 6   \n",
       "\n",
       "                     GSM1687572                     GSM1687573  \\\n",
       "0             cosmic id: 687452              cosmic id: 924100   \n",
       "1      disease state: carcinoma  disease state: adenocarcinoma   \n",
       "2     disease location: bladder     disease location: prostate   \n",
       "3  organism part: Urinary tract        organism part: Prostate   \n",
       "4                     sample: 7                    sample: 746   \n",
       "5           cell line code: 505            cell line code: 760   \n",
       "6                supplier: ATCC                 supplier: DSMZ   \n",
       "7                 affy_batch: 2                  affy_batch: 1   \n",
       "8                crna plate: 11                  crna plate: 8   \n",
       "\n",
       "                      GSM1687574                                  GSM1687575  \\\n",
       "0              cosmic id: 910924                           cosmic id: 906798   \n",
       "1  disease state: adenocarcinoma  disease state: transitional cell carcinoma   \n",
       "2      disease location: stomach                    disease location: ureter   \n",
       "3         organism part: Stomach                organism part: Urinary tract   \n",
       "4                    sample: 439                                 sample: 168   \n",
       "5            cell line code: 437                         cell line code: 151   \n",
       "6                 supplier: DSMZ                              supplier: DSMZ   \n",
       "7                  affy_batch: 1                               affy_batch: 1   \n",
       "8                  crna plate: 5                               crna plate: 2   \n",
       "\n",
       "                                   GSM1687576  \\\n",
       "0                           cosmic id: 906797   \n",
       "1  disease state: transitional cell carcinoma   \n",
       "2                   disease location: bladder   \n",
       "3                organism part: Urinary tract   \n",
       "4                                 sample: 152   \n",
       "5                         cell line code: 134   \n",
       "6                              supplier: DSMZ   \n",
       "7                               affy_batch: 1   \n",
       "8                               crna plate: 2   \n",
       "\n",
       "                                   GSM1687577  \\\n",
       "0                           cosmic id: 906797   \n",
       "1  disease state: transitional cell carcinoma   \n",
       "2                   disease location: bladder   \n",
       "3                organism part: Urinary tract   \n",
       "4                                  sample: 37   \n",
       "5                         cell line code: 134   \n",
       "6                              supplier: DSMZ   \n",
       "7                               affy_batch: 2   \n",
       "8                              crna plate: 12   \n",
       "\n",
       "                                       GSM1687578  ...  \\\n",
       "0                               cosmic id: 910922  ...   \n",
       "1  disease state: clear cell renal cell carcinoma  ...   \n",
       "2                        disease location: kidney  ...   \n",
       "3                           organism part: Kidney  ...   \n",
       "4                                     sample: 450  ...   \n",
       "5                             cell line code: 449  ...   \n",
       "6                           supplier: Unspecified  ...   \n",
       "7                                   affy_batch: 1  ...   \n",
       "8                                   crna plate: 5  ...   \n",
       "\n",
       "                            GSM1688358                     GSM1688359  \\\n",
       "0                    cosmic id: 909781              cosmic id: 909782   \n",
       "1  disease state: renal cell carcinoma  disease state: retinoblastoma   \n",
       "2             disease location: kidney       disease location: retina   \n",
       "3                organism part: Kidney             organism part: Eye   \n",
       "4                          sample: 470                    sample: 246   \n",
       "5                  cell line code: 469            cell line code: 231   \n",
       "6                      supplier: HSRRB                 supplier: ATCC   \n",
       "7                        affy_batch: 1                  affy_batch: 1   \n",
       "8                        crna plate: 5                  crna plate: 3   \n",
       "\n",
       "                      GSM1688360                         GSM1688361  \\\n",
       "0              cosmic id: 909782                  cosmic id: 909784   \n",
       "1  disease state: retinoblastoma  disease state: malignant melanoma   \n",
       "2       disease location: retina             disease location: skin   \n",
       "3             organism part: Eye                organism part: Skin   \n",
       "4                    sample: 246                        sample: 714   \n",
       "5            cell line code: 231                cell line code: 727   \n",
       "6                 supplier: ATCC                     supplier: ATCC   \n",
       "7                  affy_batch: 2                      affy_batch: 1   \n",
       "8                 crna plate: 12                      crna plate: 8   \n",
       "\n",
       "                           GSM1688362                          GSM1688363  \\\n",
       "0                   cosmic id: 909785                   cosmic id: 909785   \n",
       "1  disease state: follicular lymphoma  disease state: follicular lymphoma   \n",
       "2        disease location: lymph node        disease location: lymph node   \n",
       "3             organism part: Lymphoma             organism part: Lymphoma   \n",
       "4                         sample: 482                          sample: 49   \n",
       "5                 cell line code: 481                 cell line code: 481   \n",
       "6                      supplier: DSMZ                      supplier: DSMZ   \n",
       "7                       affy_batch: 1                       affy_batch: 2   \n",
       "8                       crna plate: 5                      crna plate: 12   \n",
       "\n",
       "                   GSM1688364                              GSM1688365  \\\n",
       "0           cosmic id: 909904                       cosmic id: 909905   \n",
       "1    disease state: carcinoma  disease state: glioblastoma multiforme   \n",
       "2  disease location: pancreas                 disease location: brain   \n",
       "3     organism part: Pancreas                    organism part: Brain   \n",
       "4                 sample: 234                              sample: 41   \n",
       "5         cell line code: 219                     cell line code: 401   \n",
       "6              supplier: DSMZ                         supplier: HSRRB   \n",
       "7               affy_batch: 1                           affy_batch: 2   \n",
       "8               crna plate: 3                          crna plate: 12   \n",
       "\n",
       "                               GSM1688366                       GSM1688367  \n",
       "0                       cosmic id: 687592                cosmic id: 909907  \n",
       "1  disease state: glioblastoma multiforme  disease state: ductal carcinoma  \n",
       "2         disease location: temporal lobe         disease location: breast  \n",
       "3                    organism part: Brain            organism part: Breast  \n",
       "4                             sample: 397                      sample: 726  \n",
       "5                     cell line code: 390              cell line code: 738  \n",
       "6                         supplier: HSRRB                   supplier: ATCC  \n",
       "7                           affy_batch: 1                    affy_batch: 1  \n",
       "8                           crna plate: 4                    crna plate: 8  \n",
       "\n",
       "[9 rows x 799 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the trait row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['!Sample_characteristics_ch1',\n",
       "       'disease state: L2 Acute Lymphoblastic Leukemia',\n",
       "       'disease state: NS Acute Lymphoblastic Leukemia',\n",
       "       'disease state: carcinoma', 'disease state: adenocarcinoma',\n",
       "       'disease state: transitional cell carcinoma',\n",
       "       'disease state: clear cell renal cell carcinoma',\n",
       "       'disease state: anaplastic carcinoma',\n",
       "       'disease state: glioblastoma multiforme',\n",
       "       'disease state: malignant melanoma',\n",
       "       'disease state: rhabdomyosarcoma',\n",
       "       'disease state: mucoepidermoid carcinoma',\n",
       "       'disease state: squamous cell carcinoma',\n",
       "       'disease state: renal cell carcinoma',\n",
       "       'disease state: neuroblastoma',\n",
       "       'disease state: Acute Lymphoblastic Leukemia',\n",
       "       'disease state: M5 acute myeloid leukemia',\n",
       "       'disease state: plasma cell myeloma',\n",
       "       'disease state: L1 Acute T-Cell Lymphoblastic Leukemia',\n",
       "       'disease state: astrocytoma',\n",
       "       'disease state: B Acute Lymphoblastic Leukemia',\n",
       "       'disease state: B cell lymphoma unspecified',\n",
       "       'disease state: papillary carcinoma',\n",
       "       'disease state: papillary transitional cell carcinoma',\n",
       "       'disease state: Burkitt lymphoma',\n",
       "       'disease state: hairy cell leukemia', 'disease state: hyperplasia',\n",
       "       'disease state: papillary ductal carcinoma',\n",
       "       'disease state: blast phase chronic myeloid leukemia',\n",
       "       'disease state: hepatocellular carcinoma',\n",
       "       'disease state: Adult T-Cell Leukemia/Lymphoma',\n",
       "       'disease state: Non-Small Cell Carcinoma',\n",
       "       'disease state: ductal carcinoma',\n",
       "       'disease state: Osteoblastic Osteosarcoma',\n",
       "       'disease state: M4 acute myeloid leukemia',\n",
       "       'disease state: follicular carcinoma',\n",
       "       'disease state: M7 acute myeloid leukemia',\n",
       "       'disease state: small cell carcinoma',\n",
       "       'disease state: large cell carcinoma',\n",
       "       'disease state: acute myeloid leukemia',\n",
       "       'disease state: primitive neuroectodermal tumor-medulloblastoma',\n",
       "       'disease state: glioma', 'disease state: desmoplastic',\n",
       "       'disease state: diffuse large B cell lymphoma',\n",
       "       'disease state: lymphoid neoplasm',\n",
       "       'disease state: malignant trichilemmal cyst',\n",
       "       'disease state: epithelial dysplasia',\n",
       "       'disease state: small cell adenocarcinoma',\n",
       "       'disease state: Micropapillary Serous Carcinoma',\n",
       "       'disease state: mucinous carcinoma',\n",
       "       'disease state: chronic lymphocytic leukemia-small lymphocytic lymphoma',\n",
       "       'disease state: chronic eosinophilic leukemia-hypereosinophilic syndrome',\n",
       "       'disease state: Ewings sarcoma-peripheral primitive neuroectodermal tumor',\n",
       "       'disease state: carcinosarcoma-malignant mesodermal mixed tumor',\n",
       "       'disease state: rhabdoid tumor',\n",
       "       'disease state: malignant fibrous histiocytoma-pleomorphic sarcoma',\n",
       "       'disease state: gliosarcoma',\n",
       "       'disease state: Mycosis Fungoides and Sezary Syndrome',\n",
       "       'disease state: Hodgkin lymphoma',\n",
       "       'disease state: M6 acute myeloid leukemia',\n",
       "       'disease state: chondrosarcoma',\n",
       "       'disease state: M3 acute myeloid leukemia',\n",
       "       'disease state: osteosarcoma', 'disease state: fibrosarcoma',\n",
       "       'disease state: mesothelioma',\n",
       "       'disease state: embryonal carcinoma',\n",
       "       'disease state: choriocarcinoma',\n",
       "       'disease state: Acute T-Cell Lymphoblastic Leukemia',\n",
       "       'disease state: anaplastic large cell lymphoma',\n",
       "       'disease state: M2 Therapy-Related Myeloid Neoplasm',\n",
       "       'disease state: granulosa cell tumor',\n",
       "       'disease state: Anaplastic Astrocytoma',\n",
       "       'disease state: undifferentiated carcinoma',\n",
       "       'disease state: Acute Leukemia of Ambiguous Lineage',\n",
       "       'disease state: giant cell carcinoma',\n",
       "       'disease state: Medullary Breast Carcinoma',\n",
       "       'disease state: sarcoma', 'disease state: tubular adenocarcinoma',\n",
       "       'disease state: L2 Acute T-Cell Lymphoblastic Leukemia',\n",
       "       'disease state: biphasic mesothelioma',\n",
       "       'disease state: Refractory Anemia with Excess Blasts',\n",
       "       'disease state: teratocarcinoma',\n",
       "       'disease state: bronchioloalveolar adenocarcinoma',\n",
       "       'disease state: adenosquamous carcinoma',\n",
       "       'disease state: adrenal cortical carcinoma',\n",
       "       'disease state: carcinoid-endocrine tumor',\n",
       "       'disease state: micropapillary carcinoma',\n",
       "       'disease state: metaplasia', 'disease state: teratoma',\n",
       "       'disease state: #N/A', 'disease state: Embryonal Rhabdomyosarcoma',\n",
       "       'disease state: clear cell carcinoma',\n",
       "       'disease state: chronic myeloid leukemia',\n",
       "       'disease state: acute leukemia',\n",
       "       'disease state: multipotential sarcoma',\n",
       "       'disease state: myelodysplastic syndrome',\n",
       "       'disease state: Wilms tumor', 'disease state: leiomyosarcoma',\n",
       "       'disease state: liposarcoma', 'disease state: synovial sarcoma',\n",
       "       'disease state: medullary carcinoma',\n",
       "       'disease state: epithelioid sarcoma',\n",
       "       'disease state: retinoblastoma',\n",
       "       'disease state: follicular lymphoma'], dtype=object)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_stage_row = clinical_data.iloc[1]\n",
    "tumor_stage_row.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine the trait row, age row, and gender row. Then implement the conversion functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.385870186Z",
     "start_time": "2024-02-27T00:23:30.351051500Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "trait_row = 0\n",
    "age_row = None\n",
    "gender_row = None\n",
    "\n",
    "def convert_trait(trait):\n",
    "    if (trait == 'primary site: endometrium'):\n",
    "        return 1  \n",
    "    else:\n",
    "        return 0  \n",
    "\n",
    "def convert_age(age_string):\n",
    "    if age_string == 'n.a.':\n",
    "        return None\n",
    "    try:\n",
    "        age = int(age_string.split(': ')[1])\n",
    "        return age\n",
    "    except (ValueError, IndexError):\n",
    "        return None\n",
    "\n",
    "def convert_gender(gender_string):\n",
    "    if (gender_string.lower() == 'sex: female' or gender_string.lower() == 'sex: f' or gender_string.lower() == 'gender: female' or gender_string.lower() == 'gender: f'):\n",
    "        return 1\n",
    "    elif (gender_string.lower() == 'sex: male' or gender_string.lower() == 'sex: m' or gender_string.lower() == 'gender: male' or gender_string.lower() == 'gender: m') :  # changeed \n",
    "        return 0\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the processed clinical data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.386191717Z",
     "start_time": "2024-02-27T00:23:30.351458802Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:218: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  clinical_df = clinical_df.applymap(convert_fn)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM886835</th>\n",
       "      <th>GSM886836</th>\n",
       "      <th>GSM886837</th>\n",
       "      <th>GSM886838</th>\n",
       "      <th>GSM886839</th>\n",
       "      <th>GSM886840</th>\n",
       "      <th>GSM886841</th>\n",
       "      <th>GSM886842</th>\n",
       "      <th>GSM886843</th>\n",
       "      <th>GSM886844</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM887742</th>\n",
       "      <th>GSM887743</th>\n",
       "      <th>GSM887744</th>\n",
       "      <th>GSM887745</th>\n",
       "      <th>GSM887746</th>\n",
       "      <th>GSM887747</th>\n",
       "      <th>GSM887748</th>\n",
       "      <th>GSM887749</th>\n",
       "      <th>GSM887750</th>\n",
       "      <th>GSM887751</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Uterine Carcinosarcoma</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 917 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GSM886835  GSM886836  GSM886837  GSM886838  GSM886839  \\\n",
       "Uterine Carcinosarcoma          0          0          0          0          0   \n",
       "\n",
       "                        GSM886840  GSM886841  GSM886842  GSM886843  GSM886844  \\\n",
       "Uterine Carcinosarcoma          0          0          0          0          0   \n",
       "\n",
       "                        ...  GSM887742  GSM887743  GSM887744  GSM887745  \\\n",
       "Uterine Carcinosarcoma  ...          0          0          0          0   \n",
       "\n",
       "                        GSM887746  GSM887747  GSM887748  GSM887749  GSM887750  \\\n",
       "Uterine Carcinosarcoma          0          0          0          0          0   \n",
       "\n",
       "                        GSM887751  \n",
       "Uterine Carcinosarcoma          0  \n",
       "\n",
       "[1 rows x 917 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_clinical_data = geo_select_clinical_features(clinical_data, TRAIT, trait_row, convert_trait, age_row=age_row,\n",
    "                                                      convert_age=convert_age, gender_row=gender_row,\n",
    "                                                      convert_gender=convert_gender)\n",
    "selected_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:58:04.978664193Z",
     "start_time": "2023-12-31T03:58:04.966117261Z"
    },
    "collapsed": false
   },
   "source": [
    "### Genetic data preprocessing and final filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the genetic data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-02-27T00:23:30.394809420Z",
     "start_time": "2024-02-27T00:23:30.351566744Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM886835</th>\n",
       "      <th>GSM886836</th>\n",
       "      <th>GSM886837</th>\n",
       "      <th>GSM886838</th>\n",
       "      <th>GSM886839</th>\n",
       "      <th>GSM886840</th>\n",
       "      <th>GSM886841</th>\n",
       "      <th>GSM886842</th>\n",
       "      <th>GSM886843</th>\n",
       "      <th>GSM886844</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM887742</th>\n",
       "      <th>GSM887743</th>\n",
       "      <th>GSM887744</th>\n",
       "      <th>GSM887745</th>\n",
       "      <th>GSM887746</th>\n",
       "      <th>GSM887747</th>\n",
       "      <th>GSM887748</th>\n",
       "      <th>GSM887749</th>\n",
       "      <th>GSM887750</th>\n",
       "      <th>GSM887751</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100009676_at</th>\n",
       "      <td>6.1161</td>\n",
       "      <td>6.2052</td>\n",
       "      <td>6.1249</td>\n",
       "      <td>6.6154</td>\n",
       "      <td>5.4236</td>\n",
       "      <td>6.0204</td>\n",
       "      <td>6.3859</td>\n",
       "      <td>6.0518</td>\n",
       "      <td>5.5014</td>\n",
       "      <td>5.4835</td>\n",
       "      <td>...</td>\n",
       "      <td>5.7775</td>\n",
       "      <td>5.5271</td>\n",
       "      <td>6.5087</td>\n",
       "      <td>5.2957</td>\n",
       "      <td>5.0551</td>\n",
       "      <td>5.6623</td>\n",
       "      <td>5.4516</td>\n",
       "      <td>5.7055</td>\n",
       "      <td>5.2289</td>\n",
       "      <td>5.2136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10000_at</th>\n",
       "      <td>8.1556</td>\n",
       "      <td>6.6152</td>\n",
       "      <td>4.5676</td>\n",
       "      <td>4.3519</td>\n",
       "      <td>6.6723</td>\n",
       "      <td>6.6316</td>\n",
       "      <td>6.1898</td>\n",
       "      <td>7.5308</td>\n",
       "      <td>6.0786</td>\n",
       "      <td>7.6018</td>\n",
       "      <td>...</td>\n",
       "      <td>4.4568</td>\n",
       "      <td>5.3949</td>\n",
       "      <td>6.7777</td>\n",
       "      <td>6.4488</td>\n",
       "      <td>6.5771</td>\n",
       "      <td>6.9336</td>\n",
       "      <td>7.0296</td>\n",
       "      <td>4.6383</td>\n",
       "      <td>5.8960</td>\n",
       "      <td>4.4004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10001_at</th>\n",
       "      <td>9.7864</td>\n",
       "      <td>9.9699</td>\n",
       "      <td>8.8720</td>\n",
       "      <td>9.1376</td>\n",
       "      <td>10.0290</td>\n",
       "      <td>9.6573</td>\n",
       "      <td>9.4910</td>\n",
       "      <td>9.3053</td>\n",
       "      <td>9.5423</td>\n",
       "      <td>8.7085</td>\n",
       "      <td>...</td>\n",
       "      <td>8.6107</td>\n",
       "      <td>9.1372</td>\n",
       "      <td>9.5185</td>\n",
       "      <td>9.0581</td>\n",
       "      <td>9.8340</td>\n",
       "      <td>9.1133</td>\n",
       "      <td>8.3857</td>\n",
       "      <td>8.5755</td>\n",
       "      <td>9.3064</td>\n",
       "      <td>9.0476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10002_at</th>\n",
       "      <td>3.7977</td>\n",
       "      <td>4.0304</td>\n",
       "      <td>3.8455</td>\n",
       "      <td>3.7085</td>\n",
       "      <td>3.6431</td>\n",
       "      <td>4.0679</td>\n",
       "      <td>3.8344</td>\n",
       "      <td>3.7223</td>\n",
       "      <td>3.9649</td>\n",
       "      <td>4.2291</td>\n",
       "      <td>...</td>\n",
       "      <td>3.7260</td>\n",
       "      <td>4.3524</td>\n",
       "      <td>4.0484</td>\n",
       "      <td>4.2962</td>\n",
       "      <td>4.0643</td>\n",
       "      <td>4.1060</td>\n",
       "      <td>3.5973</td>\n",
       "      <td>3.9060</td>\n",
       "      <td>4.7090</td>\n",
       "      <td>4.0903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10003_at</th>\n",
       "      <td>3.5458</td>\n",
       "      <td>3.8504</td>\n",
       "      <td>4.0458</td>\n",
       "      <td>3.9508</td>\n",
       "      <td>4.1589</td>\n",
       "      <td>4.1798</td>\n",
       "      <td>3.7039</td>\n",
       "      <td>3.6759</td>\n",
       "      <td>3.6180</td>\n",
       "      <td>3.6748</td>\n",
       "      <td>...</td>\n",
       "      <td>3.5464</td>\n",
       "      <td>3.7550</td>\n",
       "      <td>3.8436</td>\n",
       "      <td>3.8422</td>\n",
       "      <td>3.7768</td>\n",
       "      <td>3.8934</td>\n",
       "      <td>3.8004</td>\n",
       "      <td>4.0094</td>\n",
       "      <td>3.6699</td>\n",
       "      <td>3.8667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 917 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM886835  GSM886836  GSM886837  GSM886838  GSM886839  \\\n",
       "ID                                                                    \n",
       "100009676_at     6.1161     6.2052     6.1249     6.6154     5.4236   \n",
       "10000_at         8.1556     6.6152     4.5676     4.3519     6.6723   \n",
       "10001_at         9.7864     9.9699     8.8720     9.1376    10.0290   \n",
       "10002_at         3.7977     4.0304     3.8455     3.7085     3.6431   \n",
       "10003_at         3.5458     3.8504     4.0458     3.9508     4.1589   \n",
       "\n",
       "              GSM886840  GSM886841  GSM886842  GSM886843  GSM886844  ...  \\\n",
       "ID                                                                   ...   \n",
       "100009676_at     6.0204     6.3859     6.0518     5.5014     5.4835  ...   \n",
       "10000_at         6.6316     6.1898     7.5308     6.0786     7.6018  ...   \n",
       "10001_at         9.6573     9.4910     9.3053     9.5423     8.7085  ...   \n",
       "10002_at         4.0679     3.8344     3.7223     3.9649     4.2291  ...   \n",
       "10003_at         4.1798     3.7039     3.6759     3.6180     3.6748  ...   \n",
       "\n",
       "              GSM887742  GSM887743  GSM887744  GSM887745  GSM887746  \\\n",
       "ID                                                                    \n",
       "100009676_at     5.7775     5.5271     6.5087     5.2957     5.0551   \n",
       "10000_at         4.4568     5.3949     6.7777     6.4488     6.5771   \n",
       "10001_at         8.6107     9.1372     9.5185     9.0581     9.8340   \n",
       "10002_at         3.7260     4.3524     4.0484     4.2962     4.0643   \n",
       "10003_at         3.5464     3.7550     3.8436     3.8422     3.7768   \n",
       "\n",
       "              GSM887747  GSM887748  GSM887749  GSM887750  GSM887751  \n",
       "ID                                                                   \n",
       "100009676_at     5.6623     5.4516     5.7055     5.2289     5.2136  \n",
       "10000_at         6.9336     7.0296     4.6383     5.8960     4.4004  \n",
       "10001_at         9.1133     8.3857     8.5755     9.3064     9.0476  \n",
       "10002_at         4.1060     3.5973     3.9060     4.7090     4.0903  \n",
       "10003_at         3.8934     3.8004     4.0094     3.6699     3.8667  \n",
       "\n",
       "[5 rows x 917 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = get_genetic_data(matrix_file)\n",
    "genetic_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-12-31T03:30:41.595335164Z",
     "start_time": "2023-12-31T03:30:41.513232329Z"
    },
    "collapsed": false
   },
   "source": [
    "Check if the gene dataset requires mapping to get the gene symbols corresponding to each data row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ID': ['1_at', '10_at', '100_at', '1000_at', '10000_at'], 'ORF': ['1', '10', '100', '1000', '10000'], 'Description': ['alpha-1-B glycoprotein', 'N-acetyltransferase 2 (arylamine N-acetyltransferase)', 'adenosine deaminase', 'cadherin 2, type 1, N-cadherin (neuronal)', 'v-akt murine thymoma viral oncogene homolog 3 (protein kinase B, gamma)']}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'ORF', 'Description'], dtype='object')"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requires_gene_mapping = True\n",
    "\n",
    "if requires_gene_mapping:\n",
    "    gene_annotation = get_gene_annotation(soft_file)\n",
    "    gene_annotation_summary = preview_df(gene_annotation)\n",
    "    print(gene_annotation_summary)\n",
    "\n",
    "gene_annotation.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 162. MiB for an array with shape (21254341, 1) and data type float64",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[56], line 5\u001b[0m\n\u001b[0;32m      3\u001b[0m gene_symbol_key \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mORF\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      4\u001b[0m gene_mapping \u001b[38;5;241m=\u001b[39m get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n\u001b[1;32m----> 5\u001b[0m genetic_data \u001b[38;5;241m=\u001b[39m \u001b[43mapply_gene_mapping\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenetic_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_mapping\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\legion\\Desktop\\TAIS_2024SP\\TAIS_2024SP\\task2\\Jiayi\\utils.py:177\u001b[0m, in \u001b[0;36mapply_gene_mapping\u001b[1;34m(expression_df, mapping_df)\u001b[0m\n\u001b[0;32m    174\u001b[0m mapping_df\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mID\u001b[39m\u001b[38;5;124m'\u001b[39m, inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m    176\u001b[0m \u001b[38;5;66;03m# Merge 'mapping_df' with 'expression_df' by their indices\u001b[39;00m\n\u001b[1;32m--> 177\u001b[0m merged_df \u001b[38;5;241m=\u001b[39m \u001b[43mmapping_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexpression_df\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    179\u001b[0m \u001b[38;5;66;03m# Group by 'Gene' and calculate the mean expression values\u001b[39;00m\n\u001b[0;32m    180\u001b[0m gene_expression_df \u001b[38;5;241m=\u001b[39m merged_df\u001b[38;5;241m.\u001b[39mgroupby(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGene\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mmean()\u001b[38;5;241m.\u001b[39mdropna()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\frame.py:10744\u001b[0m, in \u001b[0;36mDataFrame.join\u001b[1;34m(self, other, on, how, lsuffix, rsuffix, sort, validate)\u001b[0m\n\u001b[0;32m  10734\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m how \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcross\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m  10735\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m merge(\n\u001b[0;32m  10736\u001b[0m             \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m  10737\u001b[0m             other,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m  10742\u001b[0m             validate\u001b[38;5;241m=\u001b[39mvalidate,\n\u001b[0;32m  10743\u001b[0m         )\n\u001b[1;32m> 10744\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m  10745\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10746\u001b[0m \u001b[43m        \u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10747\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_on\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10748\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhow\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10749\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m  10750\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m  10751\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuffixes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mlsuffix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrsuffix\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10752\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10753\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m  10754\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m  10755\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m  10756\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m on \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\merge.py:184\u001b[0m, in \u001b[0;36mmerge\u001b[1;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[0;32m    169\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    170\u001b[0m     op \u001b[38;5;241m=\u001b[39m _MergeOperation(\n\u001b[0;32m    171\u001b[0m         left_df,\n\u001b[0;32m    172\u001b[0m         right_df,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    182\u001b[0m         validate\u001b[38;5;241m=\u001b[39mvalidate,\n\u001b[0;32m    183\u001b[0m     )\n\u001b[1;32m--> 184\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\merge.py:888\u001b[0m, in \u001b[0;36m_MergeOperation.get_result\u001b[1;34m(self, copy)\u001b[0m\n\u001b[0;32m    884\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_indicator_pre_merge(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright)\n\u001b[0;32m    886\u001b[0m join_index, left_indexer, right_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_join_info()\n\u001b[1;32m--> 888\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reindex_and_concat\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    889\u001b[0m \u001b[43m    \u001b[49m\u001b[43mjoin_index\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mleft_indexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mright_indexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\n\u001b[0;32m    890\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    891\u001b[0m result \u001b[38;5;241m=\u001b[39m result\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_merge_type)\n\u001b[0;32m    893\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindicator:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\merge.py:863\u001b[0m, in \u001b[0;36m_MergeOperation._reindex_and_concat\u001b[1;34m(self, join_index, left_indexer, right_indexer, copy)\u001b[0m\n\u001b[0;32m    858\u001b[0m left\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;241m=\u001b[39m join_index\n\u001b[0;32m    860\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m right_indexer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_range_indexer(\n\u001b[0;32m    861\u001b[0m     right_indexer, \u001b[38;5;28mlen\u001b[39m(right)\n\u001b[0;32m    862\u001b[0m ):\n\u001b[1;32m--> 863\u001b[0m     rmgr \u001b[38;5;241m=\u001b[39m \u001b[43mright\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreindex_indexer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    864\u001b[0m \u001b[43m        \u001b[49m\u001b[43mjoin_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    865\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_indexer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    866\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    867\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    868\u001b[0m \u001b[43m        \u001b[49m\u001b[43monly_slice\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    869\u001b[0m \u001b[43m        \u001b[49m\u001b[43mallow_dups\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    870\u001b[0m \u001b[43m        \u001b[49m\u001b[43muse_na_proxy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    871\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    872\u001b[0m     right \u001b[38;5;241m=\u001b[39m right\u001b[38;5;241m.\u001b[39m_constructor_from_mgr(rmgr, axes\u001b[38;5;241m=\u001b[39mrmgr\u001b[38;5;241m.\u001b[39maxes)\n\u001b[0;32m    873\u001b[0m right\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;241m=\u001b[39m join_index\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\internals\\managers.py:687\u001b[0m, in \u001b[0;36mBaseBlockManager.reindex_indexer\u001b[1;34m(self, new_axis, indexer, axis, fill_value, allow_dups, copy, only_slice, use_na_proxy)\u001b[0m\n\u001b[0;32m    680\u001b[0m     new_blocks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_slice_take_blocks_ax0(\n\u001b[0;32m    681\u001b[0m         indexer,\n\u001b[0;32m    682\u001b[0m         fill_value\u001b[38;5;241m=\u001b[39mfill_value,\n\u001b[0;32m    683\u001b[0m         only_slice\u001b[38;5;241m=\u001b[39monly_slice,\n\u001b[0;32m    684\u001b[0m         use_na_proxy\u001b[38;5;241m=\u001b[39muse_na_proxy,\n\u001b[0;32m    685\u001b[0m     )\n\u001b[0;32m    686\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 687\u001b[0m     new_blocks \u001b[38;5;241m=\u001b[39m \u001b[43m[\u001b[49m\n\u001b[0;32m    688\u001b[0m \u001b[43m        \u001b[49m\u001b[43mblk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtake_nd\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    689\u001b[0m \u001b[43m            \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    690\u001b[0m \u001b[43m            \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    691\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[0;32m    692\u001b[0m \u001b[43m                \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mblk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfill_value\u001b[49m\n\u001b[0;32m    693\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    694\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    695\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mblk\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mblocks\u001b[49m\n\u001b[0;32m    696\u001b[0m \u001b[43m    \u001b[49m\u001b[43m]\u001b[49m\n\u001b[0;32m    698\u001b[0m new_axes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes)\n\u001b[0;32m    699\u001b[0m new_axes[axis] \u001b[38;5;241m=\u001b[39m new_axis\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\internals\\managers.py:688\u001b[0m, in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m    680\u001b[0m     new_blocks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_slice_take_blocks_ax0(\n\u001b[0;32m    681\u001b[0m         indexer,\n\u001b[0;32m    682\u001b[0m         fill_value\u001b[38;5;241m=\u001b[39mfill_value,\n\u001b[0;32m    683\u001b[0m         only_slice\u001b[38;5;241m=\u001b[39monly_slice,\n\u001b[0;32m    684\u001b[0m         use_na_proxy\u001b[38;5;241m=\u001b[39muse_na_proxy,\n\u001b[0;32m    685\u001b[0m     )\n\u001b[0;32m    686\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    687\u001b[0m     new_blocks \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m--> 688\u001b[0m         \u001b[43mblk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtake_nd\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    689\u001b[0m \u001b[43m            \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    690\u001b[0m \u001b[43m            \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    691\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[0;32m    692\u001b[0m \u001b[43m                \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mblk\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfill_value\u001b[49m\n\u001b[0;32m    693\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    694\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    695\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m blk \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks\n\u001b[0;32m    696\u001b[0m     ]\n\u001b[0;32m    698\u001b[0m new_axes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes)\n\u001b[0;32m    699\u001b[0m new_axes[axis] \u001b[38;5;241m=\u001b[39m new_axis\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\internals\\blocks.py:1307\u001b[0m, in \u001b[0;36mBlock.take_nd\u001b[1;34m(self, indexer, axis, new_mgr_locs, fill_value)\u001b[0m\n\u001b[0;32m   1304\u001b[0m     allow_fill \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m   1306\u001b[0m \u001b[38;5;66;03m# Note: algos.take_nd has upcast logic similar to coerce_to_target_dtype\u001b[39;00m\n\u001b[1;32m-> 1307\u001b[0m new_values \u001b[38;5;241m=\u001b[39m \u001b[43malgos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtake_nd\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1308\u001b[0m \u001b[43m    \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mallow_fill\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mallow_fill\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfill_value\u001b[49m\n\u001b[0;32m   1309\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1311\u001b[0m \u001b[38;5;66;03m# Called from three places in managers, all of which satisfy\u001b[39;00m\n\u001b[0;32m   1312\u001b[0m \u001b[38;5;66;03m#  these assertions\u001b[39;00m\n\u001b[0;32m   1313\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ExtensionBlock):\n\u001b[0;32m   1314\u001b[0m     \u001b[38;5;66;03m# NB: in this case, the 'axis' kwarg will be ignored in the\u001b[39;00m\n\u001b[0;32m   1315\u001b[0m     \u001b[38;5;66;03m#  algos.take_nd call above.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\array_algos\\take.py:117\u001b[0m, in \u001b[0;36mtake_nd\u001b[1;34m(arr, indexer, axis, fill_value, allow_fill)\u001b[0m\n\u001b[0;32m    114\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mtake(indexer, fill_value\u001b[38;5;241m=\u001b[39mfill_value, allow_fill\u001b[38;5;241m=\u001b[39mallow_fill)\n\u001b[0;32m    116\u001b[0m arr \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(arr)\n\u001b[1;32m--> 117\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_take_nd_ndarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfill_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mallow_fill\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\array_algos\\take.py:157\u001b[0m, in \u001b[0;36m_take_nd_ndarray\u001b[1;34m(arr, indexer, axis, fill_value, allow_fill)\u001b[0m\n\u001b[0;32m    155\u001b[0m     out \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mempty(out_shape, dtype\u001b[38;5;241m=\u001b[39mdtype, order\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mF\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    156\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 157\u001b[0m     out \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mempty(out_shape, dtype\u001b[38;5;241m=\u001b[39mdtype)\n\u001b[0;32m    159\u001b[0m func \u001b[38;5;241m=\u001b[39m _get_take_nd_function(\n\u001b[0;32m    160\u001b[0m     arr\u001b[38;5;241m.\u001b[39mndim, arr\u001b[38;5;241m.\u001b[39mdtype, out\u001b[38;5;241m.\u001b[39mdtype, axis\u001b[38;5;241m=\u001b[39maxis, mask_info\u001b[38;5;241m=\u001b[39mmask_info\n\u001b[0;32m    161\u001b[0m )\n\u001b[0;32m    162\u001b[0m func(arr, indexer, out, fill_value)\n",
      "\u001b[1;31mMemoryError\u001b[0m: Unable to allocate 162. MiB for an array with shape (21254341, 1) and data type float64"
     ]
    }
   ],
   "source": [
    "if requires_gene_mapping:\n",
    "    identifier_key = 'ID'\n",
    "    gene_symbol_key = 'ORF'\n",
    "    gene_mapping = get_gene_mapping(gene_annotation, identifier_key, gene_symbol_key)\n",
    "    genetic_data = apply_gene_mapping(genetic_data, gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM804806</th>\n",
       "      <th>GSM804807</th>\n",
       "      <th>GSM804808</th>\n",
       "      <th>GSM804809</th>\n",
       "      <th>GSM804810</th>\n",
       "      <th>GSM804811</th>\n",
       "      <th>GSM804812</th>\n",
       "      <th>GSM804813</th>\n",
       "      <th>GSM804814</th>\n",
       "      <th>GSM804815</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM804842</th>\n",
       "      <th>GSM804843</th>\n",
       "      <th>GSM804844</th>\n",
       "      <th>GSM804845</th>\n",
       "      <th>GSM804846</th>\n",
       "      <th>GSM804847</th>\n",
       "      <th>GSM804848</th>\n",
       "      <th>GSM804849</th>\n",
       "      <th>GSM804850</th>\n",
       "      <th>GSM804851</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>-0.428374</td>\n",
       "      <td>-1.185519</td>\n",
       "      <td>0.905210</td>\n",
       "      <td>2.229184</td>\n",
       "      <td>-0.807230</td>\n",
       "      <td>-1.049002</td>\n",
       "      <td>-0.788316</td>\n",
       "      <td>0.264672</td>\n",
       "      <td>-0.209491</td>\n",
       "      <td>-0.711295</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.267915</td>\n",
       "      <td>3.873424</td>\n",
       "      <td>2.243029</td>\n",
       "      <td>1.406519</td>\n",
       "      <td>-1.114534</td>\n",
       "      <td>-0.738209</td>\n",
       "      <td>2.528905</td>\n",
       "      <td>3.867601</td>\n",
       "      <td>-0.190160</td>\n",
       "      <td>2.753432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.112597</td>\n",
       "      <td>-2.545402</td>\n",
       "      <td>0.345880</td>\n",
       "      <td>2.294041</td>\n",
       "      <td>-1.484570</td>\n",
       "      <td>-2.047867</td>\n",
       "      <td>-1.833521</td>\n",
       "      <td>0.289472</td>\n",
       "      <td>-0.083492</td>\n",
       "      <td>-0.891364</td>\n",
       "      <td>...</td>\n",
       "      <td>0.349333</td>\n",
       "      <td>1.197180</td>\n",
       "      <td>2.109467</td>\n",
       "      <td>1.263221</td>\n",
       "      <td>-1.091778</td>\n",
       "      <td>-0.798120</td>\n",
       "      <td>2.082527</td>\n",
       "      <td>0.933188</td>\n",
       "      <td>0.148014</td>\n",
       "      <td>2.245506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.414572</td>\n",
       "      <td>1.155139</td>\n",
       "      <td>0.204161</td>\n",
       "      <td>1.504030</td>\n",
       "      <td>-0.382042</td>\n",
       "      <td>-0.231401</td>\n",
       "      <td>-0.411998</td>\n",
       "      <td>0.456773</td>\n",
       "      <td>0.943821</td>\n",
       "      <td>1.132380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.474575</td>\n",
       "      <td>0.438384</td>\n",
       "      <td>3.451386</td>\n",
       "      <td>-0.047563</td>\n",
       "      <td>-1.147372</td>\n",
       "      <td>-0.428677</td>\n",
       "      <td>-2.186289</td>\n",
       "      <td>-1.243680</td>\n",
       "      <td>-0.042116</td>\n",
       "      <td>-2.034372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-1.598132</td>\n",
       "      <td>1.704536</td>\n",
       "      <td>-1.966787</td>\n",
       "      <td>2.845671</td>\n",
       "      <td>-0.677535</td>\n",
       "      <td>-1.352631</td>\n",
       "      <td>0.799846</td>\n",
       "      <td>2.531783</td>\n",
       "      <td>-1.624713</td>\n",
       "      <td>0.691356</td>\n",
       "      <td>...</td>\n",
       "      <td>0.239463</td>\n",
       "      <td>0.176425</td>\n",
       "      <td>0.871290</td>\n",
       "      <td>0.008356</td>\n",
       "      <td>-0.008356</td>\n",
       "      <td>0.070050</td>\n",
       "      <td>2.499900</td>\n",
       "      <td>-0.278881</td>\n",
       "      <td>1.293716</td>\n",
       "      <td>1.761829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.518708</td>\n",
       "      <td>-0.038120</td>\n",
       "      <td>0.456291</td>\n",
       "      <td>0.202503</td>\n",
       "      <td>0.341105</td>\n",
       "      <td>0.115908</td>\n",
       "      <td>-0.303251</td>\n",
       "      <td>0.185458</td>\n",
       "      <td>1.257013</td>\n",
       "      <td>0.833159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.330738</td>\n",
       "      <td>0.591233</td>\n",
       "      <td>3.242414</td>\n",
       "      <td>0.461115</td>\n",
       "      <td>-0.455732</td>\n",
       "      <td>-0.314194</td>\n",
       "      <td>-1.701203</td>\n",
       "      <td>-2.183079</td>\n",
       "      <td>-0.342340</td>\n",
       "      <td>-2.171190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>1.133136</td>\n",
       "      <td>-0.275506</td>\n",
       "      <td>1.333713</td>\n",
       "      <td>-1.295064</td>\n",
       "      <td>-2.577468</td>\n",
       "      <td>2.349671</td>\n",
       "      <td>-1.714859</td>\n",
       "      <td>-3.270926</td>\n",
       "      <td>-2.733860</td>\n",
       "      <td>-1.295885</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.277940</td>\n",
       "      <td>0.084062</td>\n",
       "      <td>0.138241</td>\n",
       "      <td>1.957550</td>\n",
       "      <td>-0.441590</td>\n",
       "      <td>0.420463</td>\n",
       "      <td>-4.701589</td>\n",
       "      <td>1.435610</td>\n",
       "      <td>-1.461164</td>\n",
       "      <td>-4.537579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.220626</td>\n",
       "      <td>0.242326</td>\n",
       "      <td>0.234571</td>\n",
       "      <td>0.625795</td>\n",
       "      <td>-0.150644</td>\n",
       "      <td>-0.156697</td>\n",
       "      <td>-0.340128</td>\n",
       "      <td>0.306694</td>\n",
       "      <td>-0.270027</td>\n",
       "      <td>-0.403181</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.379437</td>\n",
       "      <td>-0.171540</td>\n",
       "      <td>-1.479160</td>\n",
       "      <td>0.156511</td>\n",
       "      <td>0.443674</td>\n",
       "      <td>-0.593189</td>\n",
       "      <td>0.641436</td>\n",
       "      <td>0.494374</td>\n",
       "      <td>-0.555147</td>\n",
       "      <td>-0.049674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>-0.634988</td>\n",
       "      <td>0.121760</td>\n",
       "      <td>-0.194830</td>\n",
       "      <td>1.395852</td>\n",
       "      <td>-1.072702</td>\n",
       "      <td>-0.492642</td>\n",
       "      <td>-0.378221</td>\n",
       "      <td>1.186319</td>\n",
       "      <td>-0.190907</td>\n",
       "      <td>-0.269548</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.017909</td>\n",
       "      <td>0.279231</td>\n",
       "      <td>0.103642</td>\n",
       "      <td>-0.334219</td>\n",
       "      <td>-0.389594</td>\n",
       "      <td>0.005596</td>\n",
       "      <td>2.668182</td>\n",
       "      <td>2.161142</td>\n",
       "      <td>1.830761</td>\n",
       "      <td>3.209167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.394825</td>\n",
       "      <td>0.374459</td>\n",
       "      <td>0.132402</td>\n",
       "      <td>0.041250</td>\n",
       "      <td>0.402375</td>\n",
       "      <td>0.242625</td>\n",
       "      <td>0.056147</td>\n",
       "      <td>-0.190490</td>\n",
       "      <td>0.399305</td>\n",
       "      <td>0.309058</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.113832</td>\n",
       "      <td>-0.182934</td>\n",
       "      <td>0.602217</td>\n",
       "      <td>-0.232841</td>\n",
       "      <td>-0.079149</td>\n",
       "      <td>-0.002652</td>\n",
       "      <td>-1.439755</td>\n",
       "      <td>-0.456133</td>\n",
       "      <td>-0.135392</td>\n",
       "      <td>0.026198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.058989</td>\n",
       "      <td>-0.414844</td>\n",
       "      <td>-0.250578</td>\n",
       "      <td>0.587149</td>\n",
       "      <td>-0.111271</td>\n",
       "      <td>0.006584</td>\n",
       "      <td>0.233355</td>\n",
       "      <td>-0.006584</td>\n",
       "      <td>-0.086750</td>\n",
       "      <td>-0.108851</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.669802</td>\n",
       "      <td>-0.270942</td>\n",
       "      <td>-2.356920</td>\n",
       "      <td>-0.634593</td>\n",
       "      <td>0.783653</td>\n",
       "      <td>-0.394218</td>\n",
       "      <td>0.223815</td>\n",
       "      <td>0.786784</td>\n",
       "      <td>0.245289</td>\n",
       "      <td>2.194536</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16561 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GSM804806  GSM804807  GSM804808  GSM804809  GSM804810  GSM804811  \\\n",
       "A1BG      -0.428374  -1.185519   0.905210   2.229184  -0.807230  -1.049002   \n",
       "A1BG-AS1   0.112597  -2.545402   0.345880   2.294041  -1.484570  -2.047867   \n",
       "A1CF      -0.414572   1.155139   0.204161   1.504030  -0.382042  -0.231401   \n",
       "A2M       -1.598132   1.704536  -1.966787   2.845671  -0.677535  -1.352631   \n",
       "A2ML1     -0.518708  -0.038120   0.456291   0.202503   0.341105   0.115908   \n",
       "...             ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A     1.133136  -0.275506   1.333713  -1.295064  -2.577468   2.349671   \n",
       "ZYG11B     0.220626   0.242326   0.234571   0.625795  -0.150644  -0.156697   \n",
       "ZYX       -0.634988   0.121760  -0.194830   1.395852  -1.072702  -0.492642   \n",
       "ZZEF1      0.394825   0.374459   0.132402   0.041250   0.402375   0.242625   \n",
       "ZZZ3       0.058989  -0.414844  -0.250578   0.587149  -0.111271   0.006584   \n",
       "\n",
       "          GSM804812  GSM804813  GSM804814  GSM804815  ...  GSM804842  \\\n",
       "A1BG      -0.788316   0.264672  -0.209491  -0.711295  ...  -0.267915   \n",
       "A1BG-AS1  -1.833521   0.289472  -0.083492  -0.891364  ...   0.349333   \n",
       "A1CF      -0.411998   0.456773   0.943821   1.132380  ...   0.474575   \n",
       "A2M        0.799846   2.531783  -1.624713   0.691356  ...   0.239463   \n",
       "A2ML1     -0.303251   0.185458   1.257013   0.833159  ...   0.330738   \n",
       "...             ...        ...        ...        ...  ...        ...   \n",
       "ZYG11A    -1.714859  -3.270926  -2.733860  -1.295885  ...  -3.277940   \n",
       "ZYG11B    -0.340128   0.306694  -0.270027  -0.403181  ...  -1.379437   \n",
       "ZYX       -0.378221   1.186319  -0.190907  -0.269548  ...  -1.017909   \n",
       "ZZEF1      0.056147  -0.190490   0.399305   0.309058  ...  -0.113832   \n",
       "ZZZ3       0.233355  -0.006584  -0.086750  -0.108851  ...  -0.669802   \n",
       "\n",
       "          GSM804843  GSM804844  GSM804845  GSM804846  GSM804847  GSM804848  \\\n",
       "A1BG       3.873424   2.243029   1.406519  -1.114534  -0.738209   2.528905   \n",
       "A1BG-AS1   1.197180   2.109467   1.263221  -1.091778  -0.798120   2.082527   \n",
       "A1CF       0.438384   3.451386  -0.047563  -1.147372  -0.428677  -2.186289   \n",
       "A2M        0.176425   0.871290   0.008356  -0.008356   0.070050   2.499900   \n",
       "A2ML1      0.591233   3.242414   0.461115  -0.455732  -0.314194  -1.701203   \n",
       "...             ...        ...        ...        ...        ...        ...   \n",
       "ZYG11A     0.084062   0.138241   1.957550  -0.441590   0.420463  -4.701589   \n",
       "ZYG11B    -0.171540  -1.479160   0.156511   0.443674  -0.593189   0.641436   \n",
       "ZYX        0.279231   0.103642  -0.334219  -0.389594   0.005596   2.668182   \n",
       "ZZEF1     -0.182934   0.602217  -0.232841  -0.079149  -0.002652  -1.439755   \n",
       "ZZZ3      -0.270942  -2.356920  -0.634593   0.783653  -0.394218   0.223815   \n",
       "\n",
       "          GSM804849  GSM804850  GSM804851  \n",
       "A1BG       3.867601  -0.190160   2.753432  \n",
       "A1BG-AS1   0.933188   0.148014   2.245506  \n",
       "A1CF      -1.243680  -0.042116  -2.034372  \n",
       "A2M       -0.278881   1.293716   1.761829  \n",
       "A2ML1     -2.183079  -0.342340  -2.171190  \n",
       "...             ...        ...        ...  \n",
       "ZYG11A     1.435610  -1.461164  -4.537579  \n",
       "ZYG11B     0.494374  -0.555147  -0.049674  \n",
       "ZYX        2.161142   1.830761   3.209167  \n",
       "ZZEF1     -0.456133  -0.135392   0.026198  \n",
       "ZZZ3       0.786784   0.245289   2.194536  \n",
       "\n",
       "[16561 rows x 46 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_data = normalize_gene_symbols_in_index(genetic_data)\n",
    "genetic_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use selected clinical data and genetic data to generate the merged data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uterine Carcinosarcoma</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM804806</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.428374</td>\n",
       "      <td>0.112597</td>\n",
       "      <td>-0.414572</td>\n",
       "      <td>-1.598132</td>\n",
       "      <td>-0.518708</td>\n",
       "      <td>-0.535701</td>\n",
       "      <td>-0.002069</td>\n",
       "      <td>-0.439697</td>\n",
       "      <td>1.063816</td>\n",
       "      <td>...</td>\n",
       "      <td>0.450995</td>\n",
       "      <td>0.924672</td>\n",
       "      <td>1.158071</td>\n",
       "      <td>-1.622256</td>\n",
       "      <td>-0.265457</td>\n",
       "      <td>1.133136</td>\n",
       "      <td>0.220626</td>\n",
       "      <td>-0.634988</td>\n",
       "      <td>0.394825</td>\n",
       "      <td>0.058989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804807</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.185519</td>\n",
       "      <td>-2.545402</td>\n",
       "      <td>1.155139</td>\n",
       "      <td>1.704536</td>\n",
       "      <td>-0.038120</td>\n",
       "      <td>-0.203013</td>\n",
       "      <td>0.038813</td>\n",
       "      <td>0.094731</td>\n",
       "      <td>-0.858164</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.860528</td>\n",
       "      <td>0.738762</td>\n",
       "      <td>-0.051341</td>\n",
       "      <td>-0.870751</td>\n",
       "      <td>0.449170</td>\n",
       "      <td>-0.275506</td>\n",
       "      <td>0.242326</td>\n",
       "      <td>0.121760</td>\n",
       "      <td>0.374459</td>\n",
       "      <td>-0.414844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804808</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.905210</td>\n",
       "      <td>0.345880</td>\n",
       "      <td>0.204161</td>\n",
       "      <td>-1.966787</td>\n",
       "      <td>0.456291</td>\n",
       "      <td>0.118227</td>\n",
       "      <td>-0.125198</td>\n",
       "      <td>-0.117670</td>\n",
       "      <td>0.191601</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.252260</td>\n",
       "      <td>0.440445</td>\n",
       "      <td>-0.285612</td>\n",
       "      <td>0.301058</td>\n",
       "      <td>0.333735</td>\n",
       "      <td>1.333713</td>\n",
       "      <td>0.234571</td>\n",
       "      <td>-0.194830</td>\n",
       "      <td>0.132402</td>\n",
       "      <td>-0.250578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804809</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.229184</td>\n",
       "      <td>2.294041</td>\n",
       "      <td>1.504030</td>\n",
       "      <td>2.845671</td>\n",
       "      <td>0.202503</td>\n",
       "      <td>-0.192439</td>\n",
       "      <td>2.072114</td>\n",
       "      <td>0.065774</td>\n",
       "      <td>-0.172029</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.855826</td>\n",
       "      <td>-0.030349</td>\n",
       "      <td>0.471311</td>\n",
       "      <td>0.735769</td>\n",
       "      <td>1.174360</td>\n",
       "      <td>-1.295064</td>\n",
       "      <td>0.625795</td>\n",
       "      <td>1.395852</td>\n",
       "      <td>0.041250</td>\n",
       "      <td>0.587149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804810</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.807230</td>\n",
       "      <td>-1.484570</td>\n",
       "      <td>-0.382042</td>\n",
       "      <td>-0.677535</td>\n",
       "      <td>0.341105</td>\n",
       "      <td>0.708834</td>\n",
       "      <td>-0.136926</td>\n",
       "      <td>0.438900</td>\n",
       "      <td>1.311179</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385150</td>\n",
       "      <td>-0.205619</td>\n",
       "      <td>-2.443651</td>\n",
       "      <td>-0.599441</td>\n",
       "      <td>0.757772</td>\n",
       "      <td>-2.577468</td>\n",
       "      <td>-0.150644</td>\n",
       "      <td>-1.072702</td>\n",
       "      <td>0.402375</td>\n",
       "      <td>-0.111271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804811</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.049002</td>\n",
       "      <td>-2.047867</td>\n",
       "      <td>-0.231401</td>\n",
       "      <td>-1.352631</td>\n",
       "      <td>0.115908</td>\n",
       "      <td>0.958485</td>\n",
       "      <td>0.205130</td>\n",
       "      <td>0.523745</td>\n",
       "      <td>0.961093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.236782</td>\n",
       "      <td>0.991536</td>\n",
       "      <td>1.121354</td>\n",
       "      <td>-0.499197</td>\n",
       "      <td>0.220960</td>\n",
       "      <td>2.349671</td>\n",
       "      <td>-0.156697</td>\n",
       "      <td>-0.492642</td>\n",
       "      <td>0.242625</td>\n",
       "      <td>0.006584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804812</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.788316</td>\n",
       "      <td>-1.833521</td>\n",
       "      <td>-0.411998</td>\n",
       "      <td>0.799846</td>\n",
       "      <td>-0.303251</td>\n",
       "      <td>1.216622</td>\n",
       "      <td>-0.644191</td>\n",
       "      <td>0.039782</td>\n",
       "      <td>0.429529</td>\n",
       "      <td>...</td>\n",
       "      <td>0.062859</td>\n",
       "      <td>-0.174568</td>\n",
       "      <td>0.051341</td>\n",
       "      <td>0.378341</td>\n",
       "      <td>0.242360</td>\n",
       "      <td>-1.714859</td>\n",
       "      <td>-0.340128</td>\n",
       "      <td>-0.378221</td>\n",
       "      <td>0.056147</td>\n",
       "      <td>0.233355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804813</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.264672</td>\n",
       "      <td>0.289472</td>\n",
       "      <td>0.456773</td>\n",
       "      <td>2.531783</td>\n",
       "      <td>0.185458</td>\n",
       "      <td>0.313938</td>\n",
       "      <td>0.135998</td>\n",
       "      <td>-0.491464</td>\n",
       "      <td>-0.250708</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.560690</td>\n",
       "      <td>-0.882830</td>\n",
       "      <td>-3.383250</td>\n",
       "      <td>1.279682</td>\n",
       "      <td>0.072676</td>\n",
       "      <td>-3.270926</td>\n",
       "      <td>0.306694</td>\n",
       "      <td>1.186319</td>\n",
       "      <td>-0.190490</td>\n",
       "      <td>-0.006584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804814</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.209491</td>\n",
       "      <td>-0.083492</td>\n",
       "      <td>0.943821</td>\n",
       "      <td>-1.624713</td>\n",
       "      <td>1.257013</td>\n",
       "      <td>0.435924</td>\n",
       "      <td>0.782212</td>\n",
       "      <td>0.704644</td>\n",
       "      <td>-0.677270</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.489612</td>\n",
       "      <td>0.285653</td>\n",
       "      <td>-0.199854</td>\n",
       "      <td>-0.677793</td>\n",
       "      <td>0.177832</td>\n",
       "      <td>-2.733860</td>\n",
       "      <td>-0.270027</td>\n",
       "      <td>-0.190907</td>\n",
       "      <td>0.399305</td>\n",
       "      <td>-0.086750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804815</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.711295</td>\n",
       "      <td>-0.891364</td>\n",
       "      <td>1.132380</td>\n",
       "      <td>0.691356</td>\n",
       "      <td>0.833159</td>\n",
       "      <td>-0.674847</td>\n",
       "      <td>0.002069</td>\n",
       "      <td>0.539719</td>\n",
       "      <td>1.217782</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.388823</td>\n",
       "      <td>0.550212</td>\n",
       "      <td>0.093551</td>\n",
       "      <td>-0.364940</td>\n",
       "      <td>0.475117</td>\n",
       "      <td>-1.295885</td>\n",
       "      <td>-0.403181</td>\n",
       "      <td>-0.269548</td>\n",
       "      <td>0.309058</td>\n",
       "      <td>-0.108851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804816</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.390606</td>\n",
       "      <td>1.282936</td>\n",
       "      <td>3.383030</td>\n",
       "      <td>-2.642929</td>\n",
       "      <td>2.648696</td>\n",
       "      <td>4.266148</td>\n",
       "      <td>1.873615</td>\n",
       "      <td>0.936284</td>\n",
       "      <td>-1.025969</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.319454</td>\n",
       "      <td>0.997735</td>\n",
       "      <td>-1.657218</td>\n",
       "      <td>-1.653960</td>\n",
       "      <td>0.125274</td>\n",
       "      <td>-0.280738</td>\n",
       "      <td>-2.429900</td>\n",
       "      <td>0.240540</td>\n",
       "      <td>1.137678</td>\n",
       "      <td>-1.233861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804817</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.710139</td>\n",
       "      <td>-1.324518</td>\n",
       "      <td>0.761841</td>\n",
       "      <td>-0.306831</td>\n",
       "      <td>1.003293</td>\n",
       "      <td>0.240304</td>\n",
       "      <td>1.680693</td>\n",
       "      <td>0.199918</td>\n",
       "      <td>-0.237229</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.120971</td>\n",
       "      <td>0.029628</td>\n",
       "      <td>1.032496</td>\n",
       "      <td>0.148143</td>\n",
       "      <td>-0.032924</td>\n",
       "      <td>3.125128</td>\n",
       "      <td>-0.353407</td>\n",
       "      <td>-0.600680</td>\n",
       "      <td>-0.010966</td>\n",
       "      <td>-0.096380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804818</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.695229</td>\n",
       "      <td>-1.173119</td>\n",
       "      <td>-0.244067</td>\n",
       "      <td>0.907819</td>\n",
       "      <td>-0.471869</td>\n",
       "      <td>1.219427</td>\n",
       "      <td>1.364358</td>\n",
       "      <td>-0.458223</td>\n",
       "      <td>-0.305900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.343450</td>\n",
       "      <td>-0.198261</td>\n",
       "      <td>-0.475754</td>\n",
       "      <td>0.400482</td>\n",
       "      <td>-0.164801</td>\n",
       "      <td>0.010023</td>\n",
       "      <td>-0.507770</td>\n",
       "      <td>0.225622</td>\n",
       "      <td>-0.089319</td>\n",
       "      <td>-0.076195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804819</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.678943</td>\n",
       "      <td>-0.156337</td>\n",
       "      <td>-0.638632</td>\n",
       "      <td>-0.573290</td>\n",
       "      <td>-0.507522</td>\n",
       "      <td>0.114590</td>\n",
       "      <td>-0.662183</td>\n",
       "      <td>-0.352016</td>\n",
       "      <td>0.663802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.613100</td>\n",
       "      <td>0.249747</td>\n",
       "      <td>1.181405</td>\n",
       "      <td>-1.371973</td>\n",
       "      <td>-0.345682</td>\n",
       "      <td>1.273870</td>\n",
       "      <td>-0.508618</td>\n",
       "      <td>-1.324914</td>\n",
       "      <td>-0.624030</td>\n",
       "      <td>0.212984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804820</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.496459</td>\n",
       "      <td>-0.800597</td>\n",
       "      <td>0.595749</td>\n",
       "      <td>-1.445540</td>\n",
       "      <td>0.019770</td>\n",
       "      <td>-0.467467</td>\n",
       "      <td>0.403535</td>\n",
       "      <td>0.742528</td>\n",
       "      <td>-0.130062</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.448057</td>\n",
       "      <td>-0.125666</td>\n",
       "      <td>-0.155693</td>\n",
       "      <td>-0.131671</td>\n",
       "      <td>0.757868</td>\n",
       "      <td>2.822178</td>\n",
       "      <td>-0.290649</td>\n",
       "      <td>-0.685346</td>\n",
       "      <td>-0.246397</td>\n",
       "      <td>0.382994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804821</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.151215</td>\n",
       "      <td>-1.664488</td>\n",
       "      <td>-0.477416</td>\n",
       "      <td>0.273552</td>\n",
       "      <td>-0.989811</td>\n",
       "      <td>-0.482698</td>\n",
       "      <td>-0.426089</td>\n",
       "      <td>-0.677159</td>\n",
       "      <td>-0.541754</td>\n",
       "      <td>...</td>\n",
       "      <td>0.322448</td>\n",
       "      <td>0.121805</td>\n",
       "      <td>0.814852</td>\n",
       "      <td>0.250706</td>\n",
       "      <td>-0.070842</td>\n",
       "      <td>2.958247</td>\n",
       "      <td>-0.228590</td>\n",
       "      <td>0.426789</td>\n",
       "      <td>-0.132319</td>\n",
       "      <td>0.722928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804822</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.144806</td>\n",
       "      <td>0.083492</td>\n",
       "      <td>0.424055</td>\n",
       "      <td>0.915587</td>\n",
       "      <td>0.358494</td>\n",
       "      <td>1.765673</td>\n",
       "      <td>0.297714</td>\n",
       "      <td>0.008301</td>\n",
       "      <td>0.469327</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.126956</td>\n",
       "      <td>-0.439043</td>\n",
       "      <td>-0.344149</td>\n",
       "      <td>0.242552</td>\n",
       "      <td>-0.211427</td>\n",
       "      <td>-0.888919</td>\n",
       "      <td>-0.609752</td>\n",
       "      <td>-0.300030</td>\n",
       "      <td>-0.483364</td>\n",
       "      <td>0.236458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804823</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.086049</td>\n",
       "      <td>-2.993872</td>\n",
       "      <td>1.464892</td>\n",
       "      <td>-0.538263</td>\n",
       "      <td>1.489029</td>\n",
       "      <td>-0.829083</td>\n",
       "      <td>2.208626</td>\n",
       "      <td>-0.216270</td>\n",
       "      <td>0.090918</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.048181</td>\n",
       "      <td>-0.440257</td>\n",
       "      <td>0.357764</td>\n",
       "      <td>0.955712</td>\n",
       "      <td>0.488899</td>\n",
       "      <td>1.583009</td>\n",
       "      <td>-0.157970</td>\n",
       "      <td>-0.226357</td>\n",
       "      <td>0.381257</td>\n",
       "      <td>0.344757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804824</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.239160</td>\n",
       "      <td>-2.717205</td>\n",
       "      <td>1.586744</td>\n",
       "      <td>-0.896581</td>\n",
       "      <td>1.286853</td>\n",
       "      <td>0.476571</td>\n",
       "      <td>0.878753</td>\n",
       "      <td>0.091504</td>\n",
       "      <td>-0.903709</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.653787</td>\n",
       "      <td>0.011096</td>\n",
       "      <td>-1.474104</td>\n",
       "      <td>0.448934</td>\n",
       "      <td>1.218168</td>\n",
       "      <td>0.502213</td>\n",
       "      <td>-0.327501</td>\n",
       "      <td>-0.322909</td>\n",
       "      <td>0.699587</td>\n",
       "      <td>-0.489933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804825</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.592178</td>\n",
       "      <td>0.975936</td>\n",
       "      <td>0.030208</td>\n",
       "      <td>-0.601298</td>\n",
       "      <td>0.479584</td>\n",
       "      <td>-2.225916</td>\n",
       "      <td>0.155777</td>\n",
       "      <td>-0.491296</td>\n",
       "      <td>0.043430</td>\n",
       "      <td>...</td>\n",
       "      <td>0.551731</td>\n",
       "      <td>-0.439228</td>\n",
       "      <td>1.111326</td>\n",
       "      <td>-0.125265</td>\n",
       "      <td>-0.581414</td>\n",
       "      <td>4.980460</td>\n",
       "      <td>0.636777</td>\n",
       "      <td>-0.336617</td>\n",
       "      <td>0.615980</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804826</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.440316</td>\n",
       "      <td>-2.891729</td>\n",
       "      <td>-0.143917</td>\n",
       "      <td>-0.869317</td>\n",
       "      <td>0.566475</td>\n",
       "      <td>-3.802507</td>\n",
       "      <td>-0.507620</td>\n",
       "      <td>-0.276853</td>\n",
       "      <td>0.458997</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056492</td>\n",
       "      <td>0.121764</td>\n",
       "      <td>-1.221214</td>\n",
       "      <td>0.083440</td>\n",
       "      <td>0.440602</td>\n",
       "      <td>-0.620730</td>\n",
       "      <td>0.046389</td>\n",
       "      <td>-0.806211</td>\n",
       "      <td>0.261524</td>\n",
       "      <td>-0.142458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804827</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.391704</td>\n",
       "      <td>-2.920563</td>\n",
       "      <td>1.217825</td>\n",
       "      <td>-0.287837</td>\n",
       "      <td>1.529358</td>\n",
       "      <td>2.151565</td>\n",
       "      <td>0.318772</td>\n",
       "      <td>0.176628</td>\n",
       "      <td>0.690347</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.138734</td>\n",
       "      <td>-0.535348</td>\n",
       "      <td>-2.978836</td>\n",
       "      <td>0.901714</td>\n",
       "      <td>0.873914</td>\n",
       "      <td>-1.257772</td>\n",
       "      <td>-0.590603</td>\n",
       "      <td>-0.901914</td>\n",
       "      <td>0.235393</td>\n",
       "      <td>0.017938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804828</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.947486</td>\n",
       "      <td>-1.966007</td>\n",
       "      <td>0.496773</td>\n",
       "      <td>-0.536419</td>\n",
       "      <td>0.476637</td>\n",
       "      <td>0.564874</td>\n",
       "      <td>-0.289421</td>\n",
       "      <td>0.037816</td>\n",
       "      <td>0.927539</td>\n",
       "      <td>...</td>\n",
       "      <td>0.542744</td>\n",
       "      <td>-0.313499</td>\n",
       "      <td>0.515259</td>\n",
       "      <td>0.401607</td>\n",
       "      <td>-0.035947</td>\n",
       "      <td>-0.162818</td>\n",
       "      <td>0.333844</td>\n",
       "      <td>-0.065670</td>\n",
       "      <td>0.027490</td>\n",
       "      <td>-0.164481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804829</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.486913</td>\n",
       "      <td>0.520850</td>\n",
       "      <td>-0.165696</td>\n",
       "      <td>0.379681</td>\n",
       "      <td>0.132478</td>\n",
       "      <td>-2.097306</td>\n",
       "      <td>0.972855</td>\n",
       "      <td>-0.331540</td>\n",
       "      <td>-0.404445</td>\n",
       "      <td>...</td>\n",
       "      <td>0.682768</td>\n",
       "      <td>0.324120</td>\n",
       "      <td>0.598722</td>\n",
       "      <td>-0.107376</td>\n",
       "      <td>-0.220175</td>\n",
       "      <td>-0.824041</td>\n",
       "      <td>0.276640</td>\n",
       "      <td>0.099597</td>\n",
       "      <td>-0.259003</td>\n",
       "      <td>-0.171322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804830</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.494610</td>\n",
       "      <td>0.341135</td>\n",
       "      <td>0.327582</td>\n",
       "      <td>-0.464571</td>\n",
       "      <td>-0.430228</td>\n",
       "      <td>-1.888709</td>\n",
       "      <td>-0.302574</td>\n",
       "      <td>-0.596609</td>\n",
       "      <td>-0.271755</td>\n",
       "      <td>...</td>\n",
       "      <td>0.682431</td>\n",
       "      <td>0.502359</td>\n",
       "      <td>0.499929</td>\n",
       "      <td>-0.319925</td>\n",
       "      <td>0.218914</td>\n",
       "      <td>0.905663</td>\n",
       "      <td>0.226767</td>\n",
       "      <td>0.405154</td>\n",
       "      <td>-0.673711</td>\n",
       "      <td>0.662017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804831</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.154453</td>\n",
       "      <td>2.098560</td>\n",
       "      <td>-0.626915</td>\n",
       "      <td>-0.012141</td>\n",
       "      <td>-0.353055</td>\n",
       "      <td>-1.451781</td>\n",
       "      <td>0.155805</td>\n",
       "      <td>-0.173156</td>\n",
       "      <td>-0.467281</td>\n",
       "      <td>...</td>\n",
       "      <td>0.964643</td>\n",
       "      <td>0.499443</td>\n",
       "      <td>0.252018</td>\n",
       "      <td>-0.110771</td>\n",
       "      <td>-0.903009</td>\n",
       "      <td>1.663458</td>\n",
       "      <td>0.161380</td>\n",
       "      <td>1.399043</td>\n",
       "      <td>-0.069165</td>\n",
       "      <td>0.477886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804832</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.310801</td>\n",
       "      <td>2.515681</td>\n",
       "      <td>0.619570</td>\n",
       "      <td>0.439519</td>\n",
       "      <td>-0.019770</td>\n",
       "      <td>-0.048056</td>\n",
       "      <td>0.700765</td>\n",
       "      <td>-0.608933</td>\n",
       "      <td>-0.211051</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.074140</td>\n",
       "      <td>0.406045</td>\n",
       "      <td>-0.486830</td>\n",
       "      <td>1.134279</td>\n",
       "      <td>-0.098571</td>\n",
       "      <td>-2.784843</td>\n",
       "      <td>0.640802</td>\n",
       "      <td>0.650824</td>\n",
       "      <td>-0.704572</td>\n",
       "      <td>-0.370144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804833</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.433587</td>\n",
       "      <td>0.418731</td>\n",
       "      <td>-1.142217</td>\n",
       "      <td>-0.584583</td>\n",
       "      <td>-0.389882</td>\n",
       "      <td>-2.938521</td>\n",
       "      <td>-0.049035</td>\n",
       "      <td>-0.626174</td>\n",
       "      <td>-0.895535</td>\n",
       "      <td>...</td>\n",
       "      <td>0.619037</td>\n",
       "      <td>-0.810250</td>\n",
       "      <td>-0.213049</td>\n",
       "      <td>0.285294</td>\n",
       "      <td>0.408258</td>\n",
       "      <td>0.434050</td>\n",
       "      <td>0.613999</td>\n",
       "      <td>0.308989</td>\n",
       "      <td>-0.286188</td>\n",
       "      <td>-0.051157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804834</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.133157</td>\n",
       "      <td>0.831940</td>\n",
       "      <td>-0.521483</td>\n",
       "      <td>0.502314</td>\n",
       "      <td>-0.303972</td>\n",
       "      <td>-0.777917</td>\n",
       "      <td>-0.462536</td>\n",
       "      <td>-0.276447</td>\n",
       "      <td>-0.250623</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.331474</td>\n",
       "      <td>0.260128</td>\n",
       "      <td>0.321089</td>\n",
       "      <td>0.248201</td>\n",
       "      <td>-0.067904</td>\n",
       "      <td>-0.010023</td>\n",
       "      <td>-0.038355</td>\n",
       "      <td>1.155498</td>\n",
       "      <td>-0.983259</td>\n",
       "      <td>-0.541571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804835</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.108337</td>\n",
       "      <td>-0.210808</td>\n",
       "      <td>-1.145994</td>\n",
       "      <td>0.537817</td>\n",
       "      <td>-0.511210</td>\n",
       "      <td>0.048056</td>\n",
       "      <td>0.484864</td>\n",
       "      <td>-0.368123</td>\n",
       "      <td>-0.007846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282319</td>\n",
       "      <td>-0.924900</td>\n",
       "      <td>-0.670668</td>\n",
       "      <td>0.104423</td>\n",
       "      <td>-0.072448</td>\n",
       "      <td>1.567791</td>\n",
       "      <td>0.435961</td>\n",
       "      <td>0.804173</td>\n",
       "      <td>-0.302328</td>\n",
       "      <td>0.172665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804836</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.058905</td>\n",
       "      <td>0.151941</td>\n",
       "      <td>0.871079</td>\n",
       "      <td>0.198332</td>\n",
       "      <td>-0.362843</td>\n",
       "      <td>-0.943148</td>\n",
       "      <td>-1.236628</td>\n",
       "      <td>-1.013508</td>\n",
       "      <td>-0.735527</td>\n",
       "      <td>...</td>\n",
       "      <td>0.995773</td>\n",
       "      <td>-0.669605</td>\n",
       "      <td>1.253273</td>\n",
       "      <td>0.187346</td>\n",
       "      <td>-0.185491</td>\n",
       "      <td>2.251003</td>\n",
       "      <td>0.168862</td>\n",
       "      <td>-0.005596</td>\n",
       "      <td>0.202087</td>\n",
       "      <td>-0.177840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804837</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.254195</td>\n",
       "      <td>-0.098369</td>\n",
       "      <td>-0.616145</td>\n",
       "      <td>2.734587</td>\n",
       "      <td>-0.427429</td>\n",
       "      <td>-2.364447</td>\n",
       "      <td>-1.174838</td>\n",
       "      <td>-1.350340</td>\n",
       "      <td>0.359030</td>\n",
       "      <td>...</td>\n",
       "      <td>0.291943</td>\n",
       "      <td>0.762170</td>\n",
       "      <td>0.425143</td>\n",
       "      <td>1.466680</td>\n",
       "      <td>-0.740152</td>\n",
       "      <td>-0.730564</td>\n",
       "      <td>1.696093</td>\n",
       "      <td>2.994437</td>\n",
       "      <td>0.010174</td>\n",
       "      <td>0.490109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804838</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.925874</td>\n",
       "      <td>0.535786</td>\n",
       "      <td>-0.087363</td>\n",
       "      <td>2.092790</td>\n",
       "      <td>-0.821037</td>\n",
       "      <td>-1.609256</td>\n",
       "      <td>-0.250885</td>\n",
       "      <td>-1.535690</td>\n",
       "      <td>-0.050420</td>\n",
       "      <td>...</td>\n",
       "      <td>0.156243</td>\n",
       "      <td>-0.011096</td>\n",
       "      <td>-0.141149</td>\n",
       "      <td>1.801681</td>\n",
       "      <td>-0.436141</td>\n",
       "      <td>-3.541426</td>\n",
       "      <td>1.605734</td>\n",
       "      <td>2.228283</td>\n",
       "      <td>-0.150764</td>\n",
       "      <td>0.442497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804839</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.491566</td>\n",
       "      <td>-0.877648</td>\n",
       "      <td>-2.027935</td>\n",
       "      <td>-0.567667</td>\n",
       "      <td>-1.068218</td>\n",
       "      <td>-0.322649</td>\n",
       "      <td>-1.344134</td>\n",
       "      <td>-0.586410</td>\n",
       "      <td>0.115810</td>\n",
       "      <td>...</td>\n",
       "      <td>1.477607</td>\n",
       "      <td>-0.037054</td>\n",
       "      <td>0.475747</td>\n",
       "      <td>-0.105539</td>\n",
       "      <td>-0.691089</td>\n",
       "      <td>0.999619</td>\n",
       "      <td>0.057841</td>\n",
       "      <td>0.374156</td>\n",
       "      <td>-0.558254</td>\n",
       "      <td>0.300191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804840</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.960238</td>\n",
       "      <td>-2.012282</td>\n",
       "      <td>0.020990</td>\n",
       "      <td>-1.272291</td>\n",
       "      <td>0.270757</td>\n",
       "      <td>1.582611</td>\n",
       "      <td>0.138080</td>\n",
       "      <td>0.695660</td>\n",
       "      <td>0.909899</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.429920</td>\n",
       "      <td>-0.694727</td>\n",
       "      <td>-1.110717</td>\n",
       "      <td>-0.491345</td>\n",
       "      <td>0.220623</td>\n",
       "      <td>0.016479</td>\n",
       "      <td>-0.508697</td>\n",
       "      <td>-1.506856</td>\n",
       "      <td>-0.170618</td>\n",
       "      <td>-0.480386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804841</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.744778</td>\n",
       "      <td>-0.684135</td>\n",
       "      <td>-0.363746</td>\n",
       "      <td>-0.179080</td>\n",
       "      <td>-0.492266</td>\n",
       "      <td>0.681176</td>\n",
       "      <td>-0.392032</td>\n",
       "      <td>0.032570</td>\n",
       "      <td>0.136582</td>\n",
       "      <td>...</td>\n",
       "      <td>0.048181</td>\n",
       "      <td>0.379861</td>\n",
       "      <td>0.351188</td>\n",
       "      <td>0.017402</td>\n",
       "      <td>0.558123</td>\n",
       "      <td>-4.197624</td>\n",
       "      <td>-0.393444</td>\n",
       "      <td>-0.498142</td>\n",
       "      <td>0.319042</td>\n",
       "      <td>0.062288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804842</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.267915</td>\n",
       "      <td>0.349333</td>\n",
       "      <td>0.474575</td>\n",
       "      <td>0.239463</td>\n",
       "      <td>0.330738</td>\n",
       "      <td>1.709726</td>\n",
       "      <td>0.190537</td>\n",
       "      <td>0.310947</td>\n",
       "      <td>-0.112710</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.166853</td>\n",
       "      <td>-0.559028</td>\n",
       "      <td>-0.825490</td>\n",
       "      <td>-0.656588</td>\n",
       "      <td>0.027825</td>\n",
       "      <td>-3.277940</td>\n",
       "      <td>-1.379437</td>\n",
       "      <td>-1.017909</td>\n",
       "      <td>-0.113832</td>\n",
       "      <td>-0.669802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804843</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.873424</td>\n",
       "      <td>1.197180</td>\n",
       "      <td>0.438384</td>\n",
       "      <td>0.176425</td>\n",
       "      <td>0.591233</td>\n",
       "      <td>-0.346614</td>\n",
       "      <td>0.128811</td>\n",
       "      <td>0.266102</td>\n",
       "      <td>0.007846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108862</td>\n",
       "      <td>1.134451</td>\n",
       "      <td>-0.336857</td>\n",
       "      <td>-0.096910</td>\n",
       "      <td>0.089705</td>\n",
       "      <td>0.084062</td>\n",
       "      <td>-0.171540</td>\n",
       "      <td>0.279231</td>\n",
       "      <td>-0.182934</td>\n",
       "      <td>-0.270942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804844</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.243029</td>\n",
       "      <td>2.109467</td>\n",
       "      <td>3.451386</td>\n",
       "      <td>0.871290</td>\n",
       "      <td>3.242414</td>\n",
       "      <td>4.464290</td>\n",
       "      <td>2.267247</td>\n",
       "      <td>1.119255</td>\n",
       "      <td>-0.920048</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.788216</td>\n",
       "      <td>0.770539</td>\n",
       "      <td>-1.887722</td>\n",
       "      <td>0.371753</td>\n",
       "      <td>-0.458492</td>\n",
       "      <td>0.138241</td>\n",
       "      <td>-1.479160</td>\n",
       "      <td>0.103642</td>\n",
       "      <td>0.602217</td>\n",
       "      <td>-2.356920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804845</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.406519</td>\n",
       "      <td>1.263221</td>\n",
       "      <td>-0.047563</td>\n",
       "      <td>0.008356</td>\n",
       "      <td>0.461115</td>\n",
       "      <td>-0.877938</td>\n",
       "      <td>-0.132084</td>\n",
       "      <td>-0.338088</td>\n",
       "      <td>0.263310</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.063442</td>\n",
       "      <td>0.870010</td>\n",
       "      <td>0.890022</td>\n",
       "      <td>-2.463706</td>\n",
       "      <td>-0.328319</td>\n",
       "      <td>1.957550</td>\n",
       "      <td>0.156511</td>\n",
       "      <td>-0.334219</td>\n",
       "      <td>-0.232841</td>\n",
       "      <td>-0.634593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804846</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.114534</td>\n",
       "      <td>-1.091778</td>\n",
       "      <td>-1.147372</td>\n",
       "      <td>-0.008356</td>\n",
       "      <td>-0.455732</td>\n",
       "      <td>-0.516802</td>\n",
       "      <td>-0.332936</td>\n",
       "      <td>-0.008301</td>\n",
       "      <td>0.426692</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130259</td>\n",
       "      <td>0.538910</td>\n",
       "      <td>1.107416</td>\n",
       "      <td>-0.942750</td>\n",
       "      <td>0.491962</td>\n",
       "      <td>-0.441590</td>\n",
       "      <td>0.443674</td>\n",
       "      <td>-0.389594</td>\n",
       "      <td>-0.079149</td>\n",
       "      <td>0.783653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804847</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.738209</td>\n",
       "      <td>-0.798120</td>\n",
       "      <td>-0.428677</td>\n",
       "      <td>0.070050</td>\n",
       "      <td>-0.314194</td>\n",
       "      <td>0.932867</td>\n",
       "      <td>-0.122364</td>\n",
       "      <td>0.035001</td>\n",
       "      <td>0.065712</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.301434</td>\n",
       "      <td>-0.194717</td>\n",
       "      <td>-0.358971</td>\n",
       "      <td>-1.337905</td>\n",
       "      <td>0.113667</td>\n",
       "      <td>0.420463</td>\n",
       "      <td>-0.593189</td>\n",
       "      <td>0.005596</td>\n",
       "      <td>-0.002652</td>\n",
       "      <td>-0.394218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804848</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.528905</td>\n",
       "      <td>2.082527</td>\n",
       "      <td>-2.186289</td>\n",
       "      <td>2.499900</td>\n",
       "      <td>-1.701203</td>\n",
       "      <td>0.342192</td>\n",
       "      <td>-1.133815</td>\n",
       "      <td>-0.068205</td>\n",
       "      <td>-0.221498</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.357454</td>\n",
       "      <td>-0.814285</td>\n",
       "      <td>1.257436</td>\n",
       "      <td>0.740486</td>\n",
       "      <td>-1.181619</td>\n",
       "      <td>-4.701589</td>\n",
       "      <td>0.641436</td>\n",
       "      <td>2.668182</td>\n",
       "      <td>-1.439755</td>\n",
       "      <td>0.223815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804849</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.867601</td>\n",
       "      <td>0.933188</td>\n",
       "      <td>-1.243680</td>\n",
       "      <td>-0.278881</td>\n",
       "      <td>-2.183079</td>\n",
       "      <td>-0.369607</td>\n",
       "      <td>-1.055983</td>\n",
       "      <td>1.106097</td>\n",
       "      <td>-0.019953</td>\n",
       "      <td>...</td>\n",
       "      <td>0.688094</td>\n",
       "      <td>-0.476535</td>\n",
       "      <td>0.422736</td>\n",
       "      <td>0.789530</td>\n",
       "      <td>-1.075626</td>\n",
       "      <td>1.435610</td>\n",
       "      <td>0.494374</td>\n",
       "      <td>2.161142</td>\n",
       "      <td>-0.456133</td>\n",
       "      <td>0.786784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804850</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.190160</td>\n",
       "      <td>0.148014</td>\n",
       "      <td>-0.042116</td>\n",
       "      <td>1.293716</td>\n",
       "      <td>-0.342340</td>\n",
       "      <td>0.524866</td>\n",
       "      <td>-0.202258</td>\n",
       "      <td>0.368563</td>\n",
       "      <td>0.146014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.256675</td>\n",
       "      <td>-1.695494</td>\n",
       "      <td>-1.434581</td>\n",
       "      <td>-0.916578</td>\n",
       "      <td>-0.647612</td>\n",
       "      <td>-1.461164</td>\n",
       "      <td>-0.555147</td>\n",
       "      <td>1.830761</td>\n",
       "      <td>-0.135392</td>\n",
       "      <td>0.245289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM804851</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.753432</td>\n",
       "      <td>2.245506</td>\n",
       "      <td>-2.034372</td>\n",
       "      <td>1.761829</td>\n",
       "      <td>-2.171190</td>\n",
       "      <td>0.456490</td>\n",
       "      <td>-1.012704</td>\n",
       "      <td>0.708047</td>\n",
       "      <td>0.286111</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.068768</td>\n",
       "      <td>-1.728762</td>\n",
       "      <td>-1.183928</td>\n",
       "      <td>0.344576</td>\n",
       "      <td>-1.654749</td>\n",
       "      <td>-4.537579</td>\n",
       "      <td>-0.049674</td>\n",
       "      <td>3.209167</td>\n",
       "      <td>0.026198</td>\n",
       "      <td>2.194536</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46 rows × 16562 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Uterine Carcinosarcoma      A1BG  A1BG-AS1      A1CF       A2M  \\\n",
       "GSM804806                     1.0 -0.428374  0.112597 -0.414572 -1.598132   \n",
       "GSM804807                     1.0 -1.185519 -2.545402  1.155139  1.704536   \n",
       "GSM804808                     1.0  0.905210  0.345880  0.204161 -1.966787   \n",
       "GSM804809                     1.0  2.229184  2.294041  1.504030  2.845671   \n",
       "GSM804810                     0.0 -0.807230 -1.484570 -0.382042 -0.677535   \n",
       "GSM804811                     0.0 -1.049002 -2.047867 -0.231401 -1.352631   \n",
       "GSM804812                     0.0 -0.788316 -1.833521 -0.411998  0.799846   \n",
       "GSM804813                     0.0  0.264672  0.289472  0.456773  2.531783   \n",
       "GSM804814                     0.0 -0.209491 -0.083492  0.943821 -1.624713   \n",
       "GSM804815                     0.0 -0.711295 -0.891364  1.132380  0.691356   \n",
       "GSM804816                     0.0  1.390606  1.282936  3.383030 -2.642929   \n",
       "GSM804817                     0.0 -0.710139 -1.324518  0.761841 -0.306831   \n",
       "GSM804818                     0.0 -0.695229 -1.173119 -0.244067  0.907819   \n",
       "GSM804819                     0.0 -0.678943 -0.156337 -0.638632 -0.573290   \n",
       "GSM804820                     0.0 -0.496459 -0.800597  0.595749 -1.445540   \n",
       "GSM804821                     0.0 -1.151215 -1.664488 -0.477416  0.273552   \n",
       "GSM804822                     0.0  0.144806  0.083492  0.424055  0.915587   \n",
       "GSM804823                     0.0 -0.086049 -2.993872  1.464892 -0.538263   \n",
       "GSM804824                     0.0 -0.239160 -2.717205  1.586744 -0.896581   \n",
       "GSM804825                     0.0  1.592178  0.975936  0.030208 -0.601298   \n",
       "GSM804826                     0.0 -1.440316 -2.891729 -0.143917 -0.869317   \n",
       "GSM804827                     0.0 -0.391704 -2.920563  1.217825 -0.287837   \n",
       "GSM804828                     0.0 -0.947486 -1.966007  0.496773 -0.536419   \n",
       "GSM804829                     1.0  1.486913  0.520850 -0.165696  0.379681   \n",
       "GSM804830                     1.0  2.494610  0.341135  0.327582 -0.464571   \n",
       "GSM804831                     0.0  3.154453  2.098560 -0.626915 -0.012141   \n",
       "GSM804832                     0.0  3.310801  2.515681  0.619570  0.439519   \n",
       "GSM804833                     1.0  1.433587  0.418731 -1.142217 -0.584583   \n",
       "GSM804834                     1.0  2.133157  0.831940 -0.521483  0.502314   \n",
       "GSM804835                     1.0  1.108337 -0.210808 -1.145994  0.537817   \n",
       "GSM804836                     1.0  1.058905  0.151941  0.871079  0.198332   \n",
       "GSM804837                     0.0  1.254195 -0.098369 -0.616145  2.734587   \n",
       "GSM804838                     0.0  0.925874  0.535786 -0.087363  2.092790   \n",
       "GSM804839                     1.0 -0.491566 -0.877648 -2.027935 -0.567667   \n",
       "GSM804840                     0.0 -0.960238 -2.012282  0.020990 -1.272291   \n",
       "GSM804841                     0.0 -0.744778 -0.684135 -0.363746 -0.179080   \n",
       "GSM804842                     0.0 -0.267915  0.349333  0.474575  0.239463   \n",
       "GSM804843                     1.0  3.873424  1.197180  0.438384  0.176425   \n",
       "GSM804844                     1.0  2.243029  2.109467  3.451386  0.871290   \n",
       "GSM804845                     1.0  1.406519  1.263221 -0.047563  0.008356   \n",
       "GSM804846                     0.0 -1.114534 -1.091778 -1.147372 -0.008356   \n",
       "GSM804847                     0.0 -0.738209 -0.798120 -0.428677  0.070050   \n",
       "GSM804848                     0.0  2.528905  2.082527 -2.186289  2.499900   \n",
       "GSM804849                     0.0  3.867601  0.933188 -1.243680 -0.278881   \n",
       "GSM804850                     0.0 -0.190160  0.148014 -0.042116  1.293716   \n",
       "GSM804851                     0.0  2.753432  2.245506 -2.034372  1.761829   \n",
       "\n",
       "              A2ML1    A4GALT     A4GNT      AAAS      AACS  ...      ZW10  \\\n",
       "GSM804806 -0.518708 -0.535701 -0.002069 -0.439697  1.063816  ...  0.450995   \n",
       "GSM804807 -0.038120 -0.203013  0.038813  0.094731 -0.858164  ... -0.860528   \n",
       "GSM804808  0.456291  0.118227 -0.125198 -0.117670  0.191601  ... -0.252260   \n",
       "GSM804809  0.202503 -0.192439  2.072114  0.065774 -0.172029  ... -0.855826   \n",
       "GSM804810  0.341105  0.708834 -0.136926  0.438900  1.311179  ... -0.385150   \n",
       "GSM804811  0.115908  0.958485  0.205130  0.523745  0.961093  ...  0.236782   \n",
       "GSM804812 -0.303251  1.216622 -0.644191  0.039782  0.429529  ...  0.062859   \n",
       "GSM804813  0.185458  0.313938  0.135998 -0.491464 -0.250708  ... -0.560690   \n",
       "GSM804814  1.257013  0.435924  0.782212  0.704644 -0.677270  ... -0.489612   \n",
       "GSM804815  0.833159 -0.674847  0.002069  0.539719  1.217782  ... -0.388823   \n",
       "GSM804816  2.648696  4.266148  1.873615  0.936284 -1.025969  ... -1.319454   \n",
       "GSM804817  1.003293  0.240304  1.680693  0.199918 -0.237229  ... -0.120971   \n",
       "GSM804818 -0.471869  1.219427  1.364358 -0.458223 -0.305900  ...  0.343450   \n",
       "GSM804819 -0.507522  0.114590 -0.662183 -0.352016  0.663802  ...  0.613100   \n",
       "GSM804820  0.019770 -0.467467  0.403535  0.742528 -0.130062  ... -0.448057   \n",
       "GSM804821 -0.989811 -0.482698 -0.426089 -0.677159 -0.541754  ...  0.322448   \n",
       "GSM804822  0.358494  1.765673  0.297714  0.008301  0.469327  ... -0.126956   \n",
       "GSM804823  1.489029 -0.829083  2.208626 -0.216270  0.090918  ... -0.048181   \n",
       "GSM804824  1.286853  0.476571  0.878753  0.091504 -0.903709  ... -1.653787   \n",
       "GSM804825  0.479584 -2.225916  0.155777 -0.491296  0.043430  ...  0.551731   \n",
       "GSM804826  0.566475 -3.802507 -0.507620 -0.276853  0.458997  ...  0.056492   \n",
       "GSM804827  1.529358  2.151565  0.318772  0.176628  0.690347  ... -1.138734   \n",
       "GSM804828  0.476637  0.564874 -0.289421  0.037816  0.927539  ...  0.542744   \n",
       "GSM804829  0.132478 -2.097306  0.972855 -0.331540 -0.404445  ...  0.682768   \n",
       "GSM804830 -0.430228 -1.888709 -0.302574 -0.596609 -0.271755  ...  0.682431   \n",
       "GSM804831 -0.353055 -1.451781  0.155805 -0.173156 -0.467281  ...  0.964643   \n",
       "GSM804832 -0.019770 -0.048056  0.700765 -0.608933 -0.211051  ... -0.074140   \n",
       "GSM804833 -0.389882 -2.938521 -0.049035 -0.626174 -0.895535  ...  0.619037   \n",
       "GSM804834 -0.303972 -0.777917 -0.462536 -0.276447 -0.250623  ... -0.331474   \n",
       "GSM804835 -0.511210  0.048056  0.484864 -0.368123 -0.007846  ...  0.282319   \n",
       "GSM804836 -0.362843 -0.943148 -1.236628 -1.013508 -0.735527  ...  0.995773   \n",
       "GSM804837 -0.427429 -2.364447 -1.174838 -1.350340  0.359030  ...  0.291943   \n",
       "GSM804838 -0.821037 -1.609256 -0.250885 -1.535690 -0.050420  ...  0.156243   \n",
       "GSM804839 -1.068218 -0.322649 -1.344134 -0.586410  0.115810  ...  1.477607   \n",
       "GSM804840  0.270757  1.582611  0.138080  0.695660  0.909899  ... -0.429920   \n",
       "GSM804841 -0.492266  0.681176 -0.392032  0.032570  0.136582  ...  0.048181   \n",
       "GSM804842  0.330738  1.709726  0.190537  0.310947 -0.112710  ... -1.166853   \n",
       "GSM804843  0.591233 -0.346614  0.128811  0.266102  0.007846  ...  0.108862   \n",
       "GSM804844  3.242414  4.464290  2.267247  1.119255 -0.920048  ... -0.788216   \n",
       "GSM804845  0.461115 -0.877938 -0.132084 -0.338088  0.263310  ... -0.063442   \n",
       "GSM804846 -0.455732 -0.516802 -0.332936 -0.008301  0.426692  ...  0.130259   \n",
       "GSM804847 -0.314194  0.932867 -0.122364  0.035001  0.065712  ... -0.301434   \n",
       "GSM804848 -1.701203  0.342192 -1.133815 -0.068205 -0.221498  ... -0.357454   \n",
       "GSM804849 -2.183079 -0.369607 -1.055983  1.106097 -0.019953  ...  0.688094   \n",
       "GSM804850 -0.342340  0.524866 -0.202258  0.368563  0.146014  ...  0.256675   \n",
       "GSM804851 -2.171190  0.456490 -1.012704  0.708047  0.286111  ... -1.068768   \n",
       "\n",
       "             ZWILCH     ZWINT      ZXDA      ZXDC    ZYG11A    ZYG11B  \\\n",
       "GSM804806  0.924672  1.158071 -1.622256 -0.265457  1.133136  0.220626   \n",
       "GSM804807  0.738762 -0.051341 -0.870751  0.449170 -0.275506  0.242326   \n",
       "GSM804808  0.440445 -0.285612  0.301058  0.333735  1.333713  0.234571   \n",
       "GSM804809 -0.030349  0.471311  0.735769  1.174360 -1.295064  0.625795   \n",
       "GSM804810 -0.205619 -2.443651 -0.599441  0.757772 -2.577468 -0.150644   \n",
       "GSM804811  0.991536  1.121354 -0.499197  0.220960  2.349671 -0.156697   \n",
       "GSM804812 -0.174568  0.051341  0.378341  0.242360 -1.714859 -0.340128   \n",
       "GSM804813 -0.882830 -3.383250  1.279682  0.072676 -3.270926  0.306694   \n",
       "GSM804814  0.285653 -0.199854 -0.677793  0.177832 -2.733860 -0.270027   \n",
       "GSM804815  0.550212  0.093551 -0.364940  0.475117 -1.295885 -0.403181   \n",
       "GSM804816  0.997735 -1.657218 -1.653960  0.125274 -0.280738 -2.429900   \n",
       "GSM804817  0.029628  1.032496  0.148143 -0.032924  3.125128 -0.353407   \n",
       "GSM804818 -0.198261 -0.475754  0.400482 -0.164801  0.010023 -0.507770   \n",
       "GSM804819  0.249747  1.181405 -1.371973 -0.345682  1.273870 -0.508618   \n",
       "GSM804820 -0.125666 -0.155693 -0.131671  0.757868  2.822178 -0.290649   \n",
       "GSM804821  0.121805  0.814852  0.250706 -0.070842  2.958247 -0.228590   \n",
       "GSM804822 -0.439043 -0.344149  0.242552 -0.211427 -0.888919 -0.609752   \n",
       "GSM804823 -0.440257  0.357764  0.955712  0.488899  1.583009 -0.157970   \n",
       "GSM804824  0.011096 -1.474104  0.448934  1.218168  0.502213 -0.327501   \n",
       "GSM804825 -0.439228  1.111326 -0.125265 -0.581414  4.980460  0.636777   \n",
       "GSM804826  0.121764 -1.221214  0.083440  0.440602 -0.620730  0.046389   \n",
       "GSM804827 -0.535348 -2.978836  0.901714  0.873914 -1.257772 -0.590603   \n",
       "GSM804828 -0.313499  0.515259  0.401607 -0.035947 -0.162818  0.333844   \n",
       "GSM804829  0.324120  0.598722 -0.107376 -0.220175 -0.824041  0.276640   \n",
       "GSM804830  0.502359  0.499929 -0.319925  0.218914  0.905663  0.226767   \n",
       "GSM804831  0.499443  0.252018 -0.110771 -0.903009  1.663458  0.161380   \n",
       "GSM804832  0.406045 -0.486830  1.134279 -0.098571 -2.784843  0.640802   \n",
       "GSM804833 -0.810250 -0.213049  0.285294  0.408258  0.434050  0.613999   \n",
       "GSM804834  0.260128  0.321089  0.248201 -0.067904 -0.010023 -0.038355   \n",
       "GSM804835 -0.924900 -0.670668  0.104423 -0.072448  1.567791  0.435961   \n",
       "GSM804836 -0.669605  1.253273  0.187346 -0.185491  2.251003  0.168862   \n",
       "GSM804837  0.762170  0.425143  1.466680 -0.740152 -0.730564  1.696093   \n",
       "GSM804838 -0.011096 -0.141149  1.801681 -0.436141 -3.541426  1.605734   \n",
       "GSM804839 -0.037054  0.475747 -0.105539 -0.691089  0.999619  0.057841   \n",
       "GSM804840 -0.694727 -1.110717 -0.491345  0.220623  0.016479 -0.508697   \n",
       "GSM804841  0.379861  0.351188  0.017402  0.558123 -4.197624 -0.393444   \n",
       "GSM804842 -0.559028 -0.825490 -0.656588  0.027825 -3.277940 -1.379437   \n",
       "GSM804843  1.134451 -0.336857 -0.096910  0.089705  0.084062 -0.171540   \n",
       "GSM804844  0.770539 -1.887722  0.371753 -0.458492  0.138241 -1.479160   \n",
       "GSM804845  0.870010  0.890022 -2.463706 -0.328319  1.957550  0.156511   \n",
       "GSM804846  0.538910  1.107416 -0.942750  0.491962 -0.441590  0.443674   \n",
       "GSM804847 -0.194717 -0.358971 -1.337905  0.113667  0.420463 -0.593189   \n",
       "GSM804848 -0.814285  1.257436  0.740486 -1.181619 -4.701589  0.641436   \n",
       "GSM804849 -0.476535  0.422736  0.789530 -1.075626  1.435610  0.494374   \n",
       "GSM804850 -1.695494 -1.434581 -0.916578 -0.647612 -1.461164 -0.555147   \n",
       "GSM804851 -1.728762 -1.183928  0.344576 -1.654749 -4.537579 -0.049674   \n",
       "\n",
       "                ZYX     ZZEF1      ZZZ3  \n",
       "GSM804806 -0.634988  0.394825  0.058989  \n",
       "GSM804807  0.121760  0.374459 -0.414844  \n",
       "GSM804808 -0.194830  0.132402 -0.250578  \n",
       "GSM804809  1.395852  0.041250  0.587149  \n",
       "GSM804810 -1.072702  0.402375 -0.111271  \n",
       "GSM804811 -0.492642  0.242625  0.006584  \n",
       "GSM804812 -0.378221  0.056147  0.233355  \n",
       "GSM804813  1.186319 -0.190490 -0.006584  \n",
       "GSM804814 -0.190907  0.399305 -0.086750  \n",
       "GSM804815 -0.269548  0.309058 -0.108851  \n",
       "GSM804816  0.240540  1.137678 -1.233861  \n",
       "GSM804817 -0.600680 -0.010966 -0.096380  \n",
       "GSM804818  0.225622 -0.089319 -0.076195  \n",
       "GSM804819 -1.324914 -0.624030  0.212984  \n",
       "GSM804820 -0.685346 -0.246397  0.382994  \n",
       "GSM804821  0.426789 -0.132319  0.722928  \n",
       "GSM804822 -0.300030 -0.483364  0.236458  \n",
       "GSM804823 -0.226357  0.381257  0.344757  \n",
       "GSM804824 -0.322909  0.699587 -0.489933  \n",
       "GSM804825 -0.336617  0.615980  0.250000  \n",
       "GSM804826 -0.806211  0.261524 -0.142458  \n",
       "GSM804827 -0.901914  0.235393  0.017938  \n",
       "GSM804828 -0.065670  0.027490 -0.164481  \n",
       "GSM804829  0.099597 -0.259003 -0.171322  \n",
       "GSM804830  0.405154 -0.673711  0.662017  \n",
       "GSM804831  1.399043 -0.069165  0.477886  \n",
       "GSM804832  0.650824 -0.704572 -0.370144  \n",
       "GSM804833  0.308989 -0.286188 -0.051157  \n",
       "GSM804834  1.155498 -0.983259 -0.541571  \n",
       "GSM804835  0.804173 -0.302328  0.172665  \n",
       "GSM804836 -0.005596  0.202087 -0.177840  \n",
       "GSM804837  2.994437  0.010174  0.490109  \n",
       "GSM804838  2.228283 -0.150764  0.442497  \n",
       "GSM804839  0.374156 -0.558254  0.300191  \n",
       "GSM804840 -1.506856 -0.170618 -0.480386  \n",
       "GSM804841 -0.498142  0.319042  0.062288  \n",
       "GSM804842 -1.017909 -0.113832 -0.669802  \n",
       "GSM804843  0.279231 -0.182934 -0.270942  \n",
       "GSM804844  0.103642  0.602217 -2.356920  \n",
       "GSM804845 -0.334219 -0.232841 -0.634593  \n",
       "GSM804846 -0.389594 -0.079149  0.783653  \n",
       "GSM804847  0.005596 -0.002652 -0.394218  \n",
       "GSM804848  2.668182 -1.439755  0.223815  \n",
       "GSM804849  2.161142 -0.456133  0.786784  \n",
       "GSM804850  1.830761 -0.135392  0.245289  \n",
       "GSM804851  3.209167  0.026198  2.194536  \n",
       "\n",
       "[46 rows x 16562 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data = geo_merge_clinical_genetic_data(selected_clinical_data, genetic_data)\n",
    "is_available = True\n",
    "\n",
    "merged_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if the merged data biased or not:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The merged dataset contains 46 samples.\n",
      "For the feature 'Uterine Carcinosarcoma', the least common label is '1.0' with 14 occurrences. This represents 30.43% of the dataset.\n",
      "The distribution of the feature 'Uterine Carcinosarcoma' in this dataset is fine.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_type = 'binary'\n",
    "print(f\"The merged dataset contains {len(merged_data)} samples.\")\n",
    "is_trait_biased, merged_data = judge_and_remove_biased_features(merged_data, TRAIT, trait_type=trait_type)\n",
    "is_trait_biased"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the data as a csv file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "if is_available:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available, is_trait_biased, merged_data, note='')\n",
    "else:\n",
    "    save_cohort_info(cohort, JSON_PATH, is_available)\n",
    "merged_data.head()\n",
    "if not is_trait_biased:\n",
    "    merged_data.to_csv(os.path.join(OUTPUT_DIR, cohort + '.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
